// @zohocharts/zc/src/core/data-processor/bubblepie-processor.js
//$Id$
/**
 *
 * @return {Array}
 */
//bubblepie-processor.js

import { DataProcessor } from '../data-processor';
import { charttype, charttypenames, chartcategory } from '../main-zc-init';
import { validateNullVal, getPropVal, pick, arrayIncludes, isObjectPropDefined } from '../../utils/zcutils';
import { isFunction } from '../../utils/function-utils';
import { ORDINAL_DATATYPE } from '../datatype-constants';
import { isMultiPieChartType } from '../../events/events-utils';
import { bubblepieUtils } from '../../bubblepie/bubblepie-utils';
import { NULL } from '../zc-constants';
import { isFacet } from '../../axis/hierarchy-axis-utils';
import { getChartTypeByIndex } from '../charttype-helpers';

DataProcessor.prototype.processBubblePieData = function(_data, options = {}) {
    // chartType = bubblepie | geobubblepie | geopie

    let chartType = options.chartType || charttype.bubblepie,
        keyAccessor = options.keyAccessor || bubblepieUtils.getKey,
        _dataIndex = options.dataIndex;

    var processor = this,
        processedseriesData = _data || processor.processedseriesData,
        chartObj = processor.chartObj,
        dataset = processor.dataset,
        commonPlotOptions = processor.data.chart.plot.plotoptions[charttypenames.get(chartType)],
        pieDataIndex = pick(_dataIndex, validateNullVal(commonPlotOptions && commonPlotOptions.dataindex, 3)),
        pieData = [],
        keys = [],
        userCategories = {
            x: NULL,
            y: []
        },
        userXCategories = getPropVal(chartObj.userdata, 'chart.axes.xaxis.categories'),
        hasUserXCategories =
            dataset.getDataType('x') == ORDINAL_DATATYPE && userXCategories && userXCategories.length && !isFacet(chartObj);

    if (hasUserXCategories) {
        userCategories.x = userXCategories;
    }

    processedseriesData.forEach(function(series, seriesIndex) {
        var yaxiscolumnorder = DataProcessor.helpers.getParsedYAxisOrder(series.yaxiscolumnorder),
            userYCategories = getPropVal(chartObj.userdata, `chart.axes.yaxis[${yaxiscolumnorder}].categories`),
            hasUserYCategories =
                dataset.getDataType('y', yaxiscolumnorder) == ORDINAL_DATATYPE &&
                userYCategories &&
                userYCategories.length &&
                !isFacet(chartObj),
            chartType = getChartTypeByIndex(chartObj, `${series.row}_${series.column}`, seriesIndex).value;

        if (hasUserYCategories) {
            userCategories.y[yaxiscolumnorder] = userYCategories;
        }

        series.data.forEach(function(subseries, subseriesIndex) {
            subseries.forEach(function(dataPoint) {
                var inactive = isInactive(dataPoint, seriesIndex, chartType);
                keys[subseriesIndex] = keys[subseriesIndex] || [];
                pieData[subseriesIndex] = pieData[subseriesIndex] || [];

                if (!inactive) {
                    var key = keyAccessor(dataPoint, seriesIndex, chartObj); //TODO: check with multi yaxiscolumnorder
                    var keyIndex = keys[subseriesIndex].indexOf(key);

                    if (keyIndex === -1) {
                        keyIndex = keys[subseriesIndex].length;
                        pieData[subseriesIndex][keyIndex] = getPieReference(dataPoint, yaxiscolumnorder);
                    }
                    pieData[subseriesIndex][keyIndex][seriesIndex] = dataPoint;
                    keys[subseriesIndex].push(key);
                }
                dataPoint.disabled = dataPoint.dummy = inactive;
            });
            pieData[subseriesIndex] = pieData[subseriesIndex] || [];
        });
    });
    addDummyDataForMissing();

    return pieData;

    // Helpers
    function isInactive(dataPoint, seriesIndex, chartType) {
        var currentSeriesData = processedseriesData[seriesIndex],
            yaxiscolumnorder = DataProcessor.helpers.getParsedYAxisOrder(currentSeriesData.yaxiscolumnorder),
            inactive = currentSeriesData.disabled || !(isMultiPieChartType(chartType) || !arrayIncludes(chartcategory.axis, chartType)),
            xVal = dataset.getX(dataPoint),
            yVal = dataset.getY(dataPoint, yaxiscolumnorder);

        if (userCategories.x) {
            inactive = inactive || userCategories.x.indexOf(xVal) === -1;
        }
        if (userCategories.y[yaxiscolumnorder]) {
            inactive = inactive || userCategories.y[yaxiscolumnorder].indexOf(yVal) === -1;
        }

        if (chartObj.isGeoMap && !inactive) {
            inactive = !processor.isValidPoint(dataPoint);
        }
        return inactive;
    }

    function addDummyDataForMissing() {
        pieData.forEach(function(subseries) {
            subseries.forEach(function(data) {
                processedseriesData.forEach(function(series, seriesIndex) {
                    if (!data[seriesIndex]) {
                        data[seriesIndex] = getDummyData(data, seriesIndex);
                    }
                });
            });
        });
    }

    function getDummyData(reference, seriesIndex) {
        var yaxiscolumnorder = DataProcessor.helpers.getParsedYAxisOrder(processedseriesData[seriesIndex].yaxiscolumnorder),
            dataset = chartObj.dataset,
            data = [];

        data[dataset.xidx] = reference.x;
        if (isFunction(dataset.yidx)) {
            data[dataset.yidx(yaxiscolumnorder)] = reference.y;
        }
        data[dataset.zidx] = reference.z;
        data[pieDataIndex] = null;
        data.dummy = true;

        return data;
    }

    function getPieReference(dataPoint, yaxiscolumnorder) {
        var meta = [];
        meta.x = dataset.getX(dataPoint);
        meta.y = dataset.getY(dataPoint, yaxiscolumnorder);
        meta.z = dataset.getZ(dataPoint);
        meta.reference = dataPoint;
        return meta;
    }
};

export var _bubblePieProcessor = DataProcessor.prototype.processBubblePieData;
