// @zohocharts/zc/src/core/data-processor/cartesian-data-processor.js
//cartesian-data-processor.js
import { d3_extent, d3_merge, d3_set, d3_ascending, d3_range, d3_map } from '../d3-constants';
import {
    defined,
    isArray,
    getPropVal,
    isObjectPropDefined,
    inherit,
    arrayIncludes,
    getDistinct,
    isWithinRange,
    objectValues,
    pick
} from '../../utils/zcutils';
import { NUMERIC_DATATYPE, TIME_DATATYPE, ORDINAL_DATATYPE } from '../datatype-constants';
import { charttype, chartcategory, charttypenames, stack } from '../main-zc-init';
import { isPointChartType, isWaterfallChartType, isBubbleChartType, isBarChartType, seriesIs } from '../../events/events-utils';
import { isAreaChartType, getUniqueChartTypes, isAlluvialChartType } from '../charttype-helpers';
// import { getStackGroupIndex } from './transforms/stack/stack-grouping';
import { DataProcessor } from '../data-processor';
import { Registry } from '../zc-registry';
import { colorscale_helpers } from '../../axis/colorscale-helpers';
import { BAR_CHARTNAME } from '../chartname-constants';
import { NULL } from '../zc-constants';
import { extendFunction } from '../../utils/function-utils';
import { getGridList } from './pivot-layout-helpers';
import { prepareStackData, AddDataToStackLayout, saveStackInfo, convertToStackPercent } from './stack-helpers';
import { FilterStore } from './filter/filter-store';

Registry.setComponent('cartesianDataProcessor', CartesianDataProcessor);
export function CartesianDataProcessor(data, dataSet, chartInstance) {
    // support variables
    this.values = [];
    this.minmaxY = [];
    this.stackeddata = [];
    this.stackedChart = null;
    this.hasNegativeValue = [];
    this.baseline = [];
    this.xRange = null;
    this.yRange = [];
    this.domainCorrections = {
        plotHeight: null,
        byDatalabels: {},
        forBubbles: {}
    };
    this.rangeCorrections = {
        datalabels: {}
    };
    this.xaxisData = data.chart.axes.xaxis;
    this.yaxesData = data.chart.axes.yaxis;
    this.qualitativeRanges = null;
    this.groupedByStack = d3_map();
    this.categoriesStore = null;

    this.hierarchical = chartInstance.preProcessor.hierarchical;
    this.categoryHierarchy = this.hierarchical && this.hierarchical.categoryTree;
    this.pivot = this.categoryHierarchy && this.hierarchical.pivot;
    this.isHierarchy = chartInstance.preProcessor.isHierarchy;
    this.hasHierarchyData = chartInstance.preProcessor.hasHierarchyData;
    chartInstance.dataObject.sankeyHierarchyLayout ? (this.sankeyHierarchyLayout = chartInstance.dataObject.sankeyHierarchyLayout) : null;

    // Super data-processor
    this.initialize = extendFunction(DataProcessor.prototype.initialize, initialize);
    DataProcessor.apply(this, arguments);
}
inherit(CartesianDataProcessor, DataProcessor);

var initialize = function(state) {
    var processor = this,
        domainCorrections = processor.domainCorrections,
        yaxesData = processor.yaxesData,
        allChartTypes = processor.allChartTypes,
        data = processor.data,
        chartObj = processor.chartObj,
        globalChartType = processor.globalChartType,
        dataSplicer = chartObj.dataSplicer,
        metadata_axes = dataSplicer.getEncode(),
        metadata_columns = dataSplicer.getColumns(),
        noofYaxis = processor.sankeyHierarchyLayout ? metadata_axes.level.length : metadata_axes.y.length,
        metadata = data.metadata,
        axes = metadata_axes,
        plotOptions = data.chart.plot.plotoptions,
        stacked = processor.isstacked(),
        dataset = processor.dataset,
        categoriesStore = (processor.categoriesStore = {}),
        processedseriesData = processor.processedseriesData,
        helpers = DataProcessor.helpers;

    const encodeName = processor.sankeyHierarchyLayout ? 'level' : 'y';

    if (processor.hierarchical && !processor.hierarchical.groupByTime && defined(state)) {
        processor.hierarchical.update();
        processor.chartObj.seriesdata = processor.hierarchical.getSeries();
    }

    //* retrieving filterStore from old dataObject
    processor.filterStore = chartObj.dataObject.filterStore ?? new FilterStore();

    processor.xRange = null;
    processor.yRange = [];
    processor.qualitativeRanges = null;
    processor.cache = {};

    domainCorrections.plotHeight = null;
    domainCorrections.forBubbles = {};

    processor.lastCategoryValue = 0;
    processor.isBubbleEnabled = false;
    processor.noofBarEnabledSeries = 0;
    processor.noofAreaEnabledSeries = 0;
    processor.noofAreaRangeEnabledSeries = 0;
    processor.noofBoxplotEnabledSeries = 0;
    processor.noofBarRangeEnabledSeries = 0;
    processor.noofGanttEnabledSeries = 0;
    processor.isContinousChart = true;
    processor.isPercentChart = false;
    const isLevelAxisCategory = processor.isLevelAxisCategory = isAlluvialChartType(chartObj);  // TODO: In future, need to change this check commonly for level axis based charts

    categoriesStore.x = [];
    categoriesStore.z = [[0]];

    categoriesStore.subseries = [];
    dataset.hasNullFormatter = {};
    processor.seriesExtermes = []; // finding series extremes for each series

    processor.seriesAssist = [];

    var yaxesGroups = chartObj.axes.y.groups;
    var yaxesleafScales = chartObj.axes.y.leaves;

    // TODO check legend filter
    //Reset cache for data
    chartObj.cache.data = {};
    chartObj.cache.search = {};
    chartObj.cache.customPoints = {};
    chartObj.cache.correction = {};

    for (var i = 0; i < noofYaxis; i++) {
        categoriesStore[`y${i}`] = [];
        processor.seriesExtermes[i] = [];
        !defined(yaxesData[i]) ? (yaxesData[i] = {}) : null;

        // Create yaxis group and mark the axis is disabled initially
        yaxesGroups[i] = yaxesGroups[i] || {};
        yaxesGroups[i].disabled = true;

        // Create yaxis group and mark the axis is disabled initially
        yaxesleafScales[i] = yaxesleafScales[i] || {};
        yaxesleafScales[i].disabled = false;
    }

    if (isArray(processedseriesData)) {
        var xColumn = isLevelAxisCategory && axes.level ? dataset.getColumnIdx('level', -1) : axes.x ? dataset.getColumn('x') : null,
            xDatatype = xColumn ? xColumn.datatype : null,
            isZenabled = defined(axes.z),
            groupbyDataIndices = helpers.groupByDataIndices(metadata, chartObj),
            check = d => arrayIncludes(allChartTypes, d);

        //sorting of data
        var xColPos = isLevelAxisCategory ? dataset.levelidx(-1) :  metadata_axes.x;
        var datatype_x = metadata_columns[xColPos].datatype;

        processor.hasZaxis = defined(dataset.zidx) && chartcategory.zaxis.some(check);
        processor.isFlowChartCategory = chartcategory.flow.some(check);
        processor.needXCategoriesSorting = false;
        var needDuplicateCategories, isPercentChart;

        var gridlist = processor.pivot ? getGridList(processor) : [{ data: processedseriesData }];
        gridlist.forEach(grid => {
            grid.data.forEach(d => {
                let disabled = d.disabled,
                    chartType_num = helpers.getChartType((d.node && d.node.type) || d.type || globalChartType, NUMERIC_DATATYPE);

                processor.isBubbleEnabled = processor.isBubbleEnabled || (!disabled && isBubbleChartType(chartType_num));
                processor.isWaterfallEnabled = processor.isWaterfallEnabled || (!disabled && isWaterfallChartType(chartType_num));
                processor.pieWithAxis = processor.pieWithAxis || (!disabled && chartType_num == charttype.bubblepie);
                processor.noofBarEnabledSeries += isBarChartType(chartType_num) && !disabled ? 1 : 0;
                processor.noofAreaEnabledSeries +=
                    (chartType_num == charttype.area || chartType_num == charttype.stkdarea) && !disabled ? 1 : 0;
                processor.noofAreaRangeEnabledSeries += chartType_num == charttype.arearange && !disabled ? 1 : 0;
                processor.noofBoxplotEnabledSeries += chartType_num == charttype.boxplot && !disabled ? 1 : 0;
                processor.noofBarRangeEnabledSeries += chartType_num == charttype.barrange && !disabled ? 1 : 0;
                processor.noofGanttEnabledSeries += chartType_num == charttype.gantt && !disabled ? 1 : 0;

                processor.isContinousChart =
                    processor.isContinousChart &&
                    (chartType_num == charttype.line || isAreaChartType(chartType_num) || isPointChartType(chartType_num));
            });
        });

        // var hooks = [];

        processedseriesData.forEach(function(d, i) {
            var disabled = d.disabled,
                curdata = d.data;
            var chartType_num = helpers.getChartType(d.type || globalChartType, NUMERIC_DATATYPE);
            var yaxiscolumnorder = helpers.getParsedYAxisOrder(d.yaxiscolumnorder);
            var yDatatype = dataset.getDataType('y', yaxiscolumnorder);

            /*
             * All size values are should be saved to encode eventhough they are disabled
             * Case: Combinational bubble chart, if all the bubble series are disabled, redraw is called, enable any bubble chart. sizeScale is not updated
             * DEMOEVENTS_BUBBLEPIE_4
             */
            if (isArray(curdata)) {
                if (defined(dataset.zidx) && arrayIncludes(chartcategory.zaxis, chartType_num)) {
                    processor.saveEncode('z', 0, i);
                }
            }

            if (!disabled && isArray(curdata)) {
                categoriesStore[`y${yaxiscolumnorder}`].enabled = true;
                yaxesGroups[yaxiscolumnorder].disabled = false;
                needDuplicateCategories = chartType_num === charttype.waterfall && needDuplicateCategories == null;
                processor.seriesAssist[i] = {};

                var actXdatatype = isLevelAxisCategory ? dataset.getDataType('level', -1, allChartTypes) :  dataset.getDataType('x', null, allChartTypes),
                    yColumn = dataset.getColumn(encodeName, yaxiscolumnorder);

                // debugger
                if (!defined(dataset.hasNullFormatter.x)) {
                    dataset.hasNullFormatter.x = !(actXdatatype === ORDINAL_DATATYPE
                        ? !isObjectPropDefined(xColumn[xDatatype], 'nullformatting')
                        : true);
                }

                if (!defined(dataset.hasNullFormatter[`y${yaxiscolumnorder}`])) {
                    dataset.hasNullFormatter[`y${yaxiscolumnorder}`] = !(yDatatype === ORDINAL_DATATYPE
                        ? !isObjectPropDefined(yColumn[yColumn.datatype], 'nullformatting')
                        : true);
                }

                processor.saveEncode('x', 0, i);
                processor.saveEncode('y', yaxiscolumnorder, i);

                if (colorscale_helpers.isQuantileColorScale(chartObj)) {
                    processor.saveEncode('color', 0, i);
                }

                var _stacked = processor.isstacked(chartType_num),
                    _percent = seriesIs(charttypenames.get(chartType_num), BAR_CHARTNAME) && plotOptions.bar.showAs;

                if (_stacked || _percent) {
                    var stackIndex;
                    if (!processor.pivot) {
                        stackIndex = stack.getStackGroupIndex(processor.getStackGroup(), i);
                        processor.saveEncode('stack', stackIndex, i);
                        processor.seriesAssist[i].layout = ['stack'];
                        processor.seriesAssist[i].stackIndex = stackIndex;

                        var stackTransform = processor.getEncode('stack', stackIndex),
                            stackLayout = Registry.getComponent('stackLayout');

                        stackTransform.layout = stackTransform.layout || new stackLayout();

                        //TODO: remove
                        saveStackInfo(processor, curdata, i, stackIndex, chartType_num);
                    }
                }

                isPercentChart = isPercentChart || _percent;

                //For stacked data need to iterate each data for collecting X and groupbyval
                //Categories store update
                processor.storeCategories(
                    curdata,
                    stacked,
                    groupbyDataIndices,
                    chartType_num,
                    actXdatatype,
                    yDatatype,
                    yaxiscolumnorder,
                    needDuplicateCategories
                );

                //Numeric store extremes update
                processor.setExtremes(curdata, stacked, yDatatype, chartType_num, isZenabled, yaxiscolumnorder);
            }
        });

        var userCategories = getPropVal(chartObj.userdata, 'chart.axes.xaxis.categories');

        if (
            processor.isAxisCategory &&
            datatype_x !== ORDINAL_DATATYPE &&
            processor.needXCategoriesSorting &&
            (!defined(userCategories) || (defined(userCategories) && userCategories.length === 0))
        ) {
            categoriesStore.x.sort(d3_ascending);
        }

        processor.isCombinationalChart = processor.pivot
            ? getUniqueChartTypes(chartObj).values().length > 1
            : getDistinct(allChartTypes, d => charttypenames.get(d)).length > 1;
        processor.isStackedPercentChart = stacked && processor.isStackedPercent();
        processor.isPercentChart = processor.isStackedPercentChart || isPercentChart;

        if (processor.pieWithAxis) {
            processor.bubblepieData = processor.processBubblePieData();
        } else if (processor.isFlowChartCategory) {
            processor.processSankeyData();
        }

        if ((stacked || processor.isPercentChart) && !processor.isWaterfallEnabled) {
            if (processor.pivot) {
                var pivotLayout = processor.hierarchical.pivotLayout,
                    rowList = pivotLayout.getRootNodeList('row'),
                    columnList = pivotLayout.getRootNodeList('column'),
                    layers = pivotLayout.getRootNodeList('layer'),
                    panes = layers.panes,
                    hasMultiCluster = rowList.hasMultiCluster(),
                    firstColumnCategoryCount = columnList.getBranchesCount(0) || 1,
                    firstRowCategoryCount = rowList.getBranchesCount(0) || 1,
                    gridlist = objectValues(panes);
                gridlist.forEach((d, i) => {
                    d.data.forEach((s, j) => {
                        var stackIndex = `${d.row}_${d.column}`,
                            seriesIndex = i + j;
                        processor.saveEncode('stack', stackIndex, i, s.node);
                        processor.seriesAssist[seriesIndex] = processor.seriesAssist[seriesIndex] || {};
                        processor.seriesAssist[seriesIndex].layout = ['stack'];
                        processor.seriesAssist[seriesIndex].stackIndex = stackIndex;

                        var stackTransform = processor.getEncode('stack', stackIndex),
                            stackLayout = Registry.getComponent('stackLayout');

                        stackTransform.layout = stackTransform.layout || new stackLayout();

                        //TODO: remove
                        var chartType_num = helpers.getChartType(d.data.type || globalChartType, NUMERIC_DATATYPE);
                        saveStackInfo(processor, d.data, seriesIndex, stackIndex, chartType_num);
                    });

                    prepareStackData(d.data, dataset, processor); //TODO: revamp function
                    AddDataToStackLayout(d.data, dataset, processor);
                });
            } else {
                prepareStackData(processedseriesData, dataset, processor); //TODO: revamp function
                AddDataToStackLayout(processedseriesData, dataset, processor);
            }
        }

        if (processor.isPercentChart) {
            var stackedData = processor.getEncode('stack');
            convertToStackPercent(stackedData, dataset);
        }
    }
};

var CartesianDataProcessorProto = CartesianDataProcessor.prototype;

/**
 *
 * @return {Array}
 */

CartesianDataProcessorProto.getSeriesData = function() {
    var processor = this,
        chart = this.chartObj;

    if (defined(processor.hierarchical) && processor.hierarchical.categories) {
        // should be updated in the chartObj.seriesdata;
        var temp = processor.hierarchical.getSeries(); //FEATURESEVENTS_SUBSERIES_1 throws error on legend filter
        chart.seriesdata = temp;
        return temp;
    }
    return chart.seriesdata;
};

/**
 *
 * @param {Number} index
 * @returns {*}
 */

CartesianDataProcessorProto.containsNegativeData = function(index) {
    return this.hasNegativeValue[index];
};

/**
 * @param  {Number} index  [description]
 * @param  {Array} extent [description]
 */
CartesianDataProcessorProto.saveBaseLine = function(index, extent, userDefinedMinRange, rangeCheck) {
    var processor = this,
        dataType = processor.dataset.getDataType('y', index),
        temp = NULL;

    if (dataType === NUMERIC_DATATYPE) {
        var baseLineValue = processor.baseline[index],
            stacked = processor.isstacked() || processor.isPercentChart;

        temp =
            defined(baseLineValue) && !stacked && (!rangeCheck || isWithinRange(baseLineValue, extent)) //On zooming only, isWithinRange should be considered
                ? baseLineValue
                : userDefinedMinRange < 0 || extent[0] < 0
                ? 0
                : NULL;
    }

    processor.hasNegativeValue[index] = temp;
};

/**
 * @returns {Array}
 */

CartesianDataProcessorProto.getMinMaxZ = function() {
    return d3_extent(d3_merge(this.categoriesStore.z));
};

/**
 *
 * @param {Array} seriesdata series points
 * @param {Array} xCategories
 * @param {Array|null} yCategories
 * @param {Number} yaxiscolumnorder
 * @param {Boolean} removeDuplicateX if true, will remove all duplicate X categories will be ignored.
 * @return {Array}
 */
CartesianDataProcessorProto.updateSeriesdatabyCategories = function(
    seriesdata,
    xCategories,
    yCategories,
    yaxiscolumnorder,
    removeDuplicateX,
    subseriesIdx
) {
    var dataSet = this.dataset,
        chart = this.chartObj,
        helpers = DataProcessor.helpers,
        isCatNullX = helpers.isCatNull(dataSet, 'x'),
        isCatNullY = helpers.isCatNull(dataSet, 'y', yaxiscolumnorder),
        tempSet = d3_set(),
        counter = 0,
        chartName = chart.chartTypes[0].name,
        // for gantt chart allow by default
        isGantt = chartName === 'gantt',
        allowDuplicateCategories = pick(chart.systemConf.chart.plot.allowDuplicateCategories, isGantt);

    // DO NOT FILTER BY CATEGORIES
    if (allowDuplicateCategories) {
        return seriesdata;
    }

    seriesdata = seriesdata.filter(function(d) {
        var show = true,
            x = dataSet.getX(d);

        if (xCategories) {
            show = show && !isCatNullX(x) && xCategories.indexOf(x) > -1;
        }

        if (yCategories) {
            var y = dataSet.getY(d, yaxiscolumnorder);
            show = show && !isCatNullY(y) && yCategories.indexOf(y) > -1;
        }

        //duplicate keys remove
        if (xCategories && removeDuplicateX && show) {
            if (!tempSet.has(x)) {
                // add the X value if does not exist
                tempSet.add(x);
            } else {
                // if the data is already in the X category, ignoring that.
                show = false;
            }
        }

        if (show) {
            d._filter = false;
            d.counter = counter++;
        } else {
            d._filter = true;
        }
        d.subseriesIdx = subseriesIdx || 0;
        return show;
    });
    return seriesdata;
};

/**
 *
 * @param curdata
 * @param stacked
 * @param yDatatype
 * @param chartType
 * @param isZenabled
 * @param yaxiscolumnorder
 */

CartesianDataProcessorProto.setExtremes = function(curdata, stacked, yDatatype, cType, isZenabled, yaxiscolumnorder) {
    var processor = this,
        dataset = processor.dataset,
        zCat = processor.categoriesStore.z;

    if (((yDatatype === NUMERIC_DATATYPE || yDatatype === TIME_DATATYPE) && !stacked) || isZenabled) {
        if (!curdata.disabled && isArray(curdata)) {
            var mergedSubgroup = d3_merge(curdata);
            if (isZenabled) {
                var zExtremes = d3_extent(mergedSubgroup, function(d) {
                    return dataset.getZ(d);
                });
                zCat.push(zExtremes);
            }
            if (yDatatype === NUMERIC_DATATYPE || yDatatype === TIME_DATATYPE) {
                var yExtremes;
                if (chartcategory.rangePlot.has(cType)) {
                    if (mergedSubgroup) {
                        var mergedYValues = mergedSubgroup.map(function(d) {
                            return dataset.getY(d, yaxiscolumnorder);
                        });
                        var mergedYranges = mergedYValues.length && mergedYValues[0] ? d3_merge(mergedYValues) : [];
                        yExtremes = d3_extent(mergedYranges);
                    } else {
                        yExtremes = d3_extent([]);
                    }
                } else {
                    yExtremes = d3_extent(mergedSubgroup, function(d) {
                        return dataset.getY(d, yaxiscolumnorder);
                    });
                }
                processor.seriesExtermes[yaxiscolumnorder].push(yExtremes);
            }
        }
    }
};

/**
 *
 * @param {Array} categories
 * @param {Boolean} leaf
 * @return {*}
 */

CartesianDataProcessorProto.getDomain = function(categories, type, leaf) {
    // var isHierarchicalMultiAxis = type == 'x' ? this.hierarchical && this.hierarchical.columns.count  : this.hierarchical && this.hierarchical.rows.count;
    let rowCount, columnCount;
    if (this.hierarchical) {
        rowCount = this.hierarchical.pivotLayout.getRootNodeList('row').getCount();
        columnCount = this.hierarchical.pivotLayout.getRootNodeList('column').getCount();
    }
    var isHierarchicalMultiAxis = type == 'x' ? columnCount : rowCount;
    if ((this.hierarchical && isHierarchicalMultiAxis && !(this.hierarchical.groupByDiscreteTime && leaf)) || this.isWaterfallEnabled) {
        //TODO waterfall
        return d3_range(categories.length);
    } else {
        return categories;
    }
};

export * from './minMaxXValues';
export * from './minMaxYValues';
export * from './categories';
export * from './getXY';
export * from './getRect';
export * from './findXCategories';
export * from './storeCategories';
export * from './stack-check';
export * from './stack-percent-check';
export * from './stack-group-list';
