// @zohocharts/zc/src/core/data-processor/flatten-data.js
/**
 *
 * @param d
 * @param groupbyDataIndices
 * @param chartType
 * @return {*|Array}
 */
import { DataProcessor } from '../data-processor';
import { Registry } from '../zc-registry';
import { chartcategory } from '../main-zc-init';
import { defined, each, validateNullVal, toCapitalize, objectValues, arrayIncludes, isArray, pick } from '../../utils/zcutils';
import { invokeFunction, isObjectPropDefined } from '../../utils/zcutils';
import { d3_map, d3_set } from '../d3-constants';
import { getEventConf, isBarChartType } from '../../events/events-utils';
import { tooltip_helpers } from '../../tooltip/tooltip-helpers';
import { isFunction } from '../../utils/function-utils';
import { getChartTypeByIndex } from '../charttype-helpers';

DataProcessor.prototype.flattenData = function(customdata) {
    let processor = this,
        chart = processor.chartObj,
        isDynamic = defined(customdata),
        chartdata = chart.dataObject.processedseriesData,
        data = getDataToFlatten(customdata, chart),
        flatten = (this.flattenPoints = this.flattenPoints || d3_map()),
        root,
        helpers = DataProcessor.helpers,
        globalChartType = processor.globalChartType,
        isCombo = processor.isCombinationalChart,
        tooltipShared = tooltip_helpers.hasSharedSeries(chart),
        xAxisReversed = chart.systemConf.chart.axes.xaxis.reversed,
        pivot = processor.pivot,
        hierarchical = processor.hierarchical;

    let groupby = 'charttype'; //or searchModel;
    //Flatten seriesdata and assign values
    var renderCounter = 0;
    data.forEach((grid, gridIdx) => {
        let paneid = `${grid.row || 0}_${grid.column || 0}`,
            gridId = `pane${grid.row || 0}_${grid.column || 0}`,
            paneIdx = `pane${grid.index || 0}`;
        if (!flatten.has(gridId)) {
            flatten.set(gridId, d3_map());
        }
        root = flatten.get(gridId);

        grid.data.forEach((series, i) => {
            let seriesIdx = isDynamic ? series.seriesindex : i,
                cType_Obj = getChartTypeByIndex(chart, paneid, seriesIdx),
                chartName = cType_Obj.name,
                rendererIdx_pivot = isDynamic ? series.rendererIndex : renderCounter++,
                rendererIdx = chartName == 'bubblepie' ? chart.bubblePieRendererIndex : pivot ? rendererIdx_pivot : seriesIdx,
                renderer = chart.renderer[rendererIdx],
                rp = renderer.commonRendererProp().params,
                isBarChart = isBarChartType(cType_Obj.value),
                chartName_alt = checkSegmentSelection(chart, renderer, chartName),
                key = getJoinbyKey(chartName_alt, groupby),
                searchModel = root.get(key) || {},
                yaxisOrder = helpers.getParsedYAxisOrder(pivot && !isDynamic ? series.node.yaxiscolumnorder : series.yaxiscolumnorder),
                chartType_num = cType_Obj.value,
                _stacked = processor.isstacked(chartType_num),
                highlighter = renderer.getHighlighter && renderer.getHighlighter(chart, seriesIdx, chartName), // TODO: while non axis chart also using event rect, renderer.getHighlighter conditon has to be removed
                seriesdata = series.data || series;

            if (!arrayIncludes(chartcategory.singleSourceEvent, chartType_num)) {
                // TODO: while non axis chart also using event rect, remove this condition
                return;
            }

            let flatted;
            if (arrayIncludes(chartcategory.hierarchy, cType_Obj.value)) {
                flatted = [];
                flatHierarchyData(seriesdata, flatted);
                seriesdata.isParent = true;
                if(chart.eventHandler.parentEventDisabled(seriesdata)){
                    seriesdata.isParent = false;
                }else{
                    flatted.push(seriesdata)
                }
            }

            each(flatted ? [flatted] : seriesdata, function(subseries, j) {
                var subseriesIdx = isDynamic ? series.subseriesindex : j;

                var id = `${seriesIdx}_${subseriesIdx}`;
                var flattenPoints = searchModel[id] || [];
                var lastItemPos = isDynamic
                        ? flattenPoints[flattenPoints.length - 1] && flattenPoints[flattenPoints.length - 1].itemIndex + 1
                            ? flattenPoints[flattenPoints.length - 1].itemIndex + 1
                            : 0
                        : 0,
                    counter = 0;

                // hash each point
                each(subseries, function(point, k) {
                    let itemIdx = lastItemPos++,
                        sortedItemIdx = itemIdx,
                        x_uniqueKeys = d3_set();
                    if (xAxisReversed) {
                        sortedItemIdx = subseries.length - 1 - sortedItemIdx;
                    }
                    var allowHash = true;
                    if (allowHash) {
                        // set index on point
                        point.seriesIndex = seriesIdx;
                        point.subSeriesIndex = subseriesIdx;
                        point.rendererIndex = rendererIdx;
                        // point.counter = counter;
                        point.yaxisOrder = yaxisOrder;
                        point._data = dataHash(chart, renderer, point, yaxisOrder, _stacked);
                        // point._key = `${chart.id}_${seriesIndex}_${subSeriesIndex}_${itemIndex}`;
                        point.paneid = paneid;
                        point._key = DataProcessor.helpers.generatePointKey(chart.id, gridId, seriesIdx, subseriesIdx, itemIdx);
                        if(!point.isParent){
                            point.itemIndex = itemIdx;
                            point.sortedItemIndex = itemIdx;
                            point._xid = chart.dataset.getActualX(point);
                            point.subseriesLength = seriesdata.length;
                        }
                        flattenPoints.push(point);
                        //Flatten levelmarker details
                        if (highlighter.levelmarker && !searchModel[`lm_${point._xid}`]) {
                            let collections = highlighter.levelmarker(point);
                            searchModel[`${chart.id}_` + `lm_${point._xid}`] = collections;
                            searchModel[`${chart.id}_` + `lm_${point._xid}`].disabled = function() {
                                return chartdata[seriesIdx].disabled || eventDisabled(chartdata[seriesIdx], chartName, chart);
                            };
                        }
                    }
                });

                searchModel[id] = flattenPoints;
                if (chartName == 'bar' || chartName == 'barrange' || chartName == 'boxplot') {
                    if (rp.whiskers && rp.whiskers.enabled && rp.whiskers.showOn == 'overlay' && rp.whiskers.events.enabled) {
                        searchModel.superKey = `${chartName}_segment`;
                    }
                }
                searchModel[id].disabled = function() {
                    let currentdata = customdata
                        ? chartdata[seriesIdx] || chartdata[i]
                        : (grid.data[seriesIdx] && grid.data[seriesIdx].node) || grid.data[seriesIdx] || grid.data[i];
                    return currentdata.disabled || eventDisabled(currentdata, chartName, chart);
                };
            });
            root.set(key, searchModel);
        });
    });

    //Flatten threshold details
    if (chart.cache.thresholdCustomPoints && !customdata) {
        for (var key in chart.cache.thresholdCustomPoints) {
            let thresholdEventObj = chart.cache.thresholdCustomPoints[key];
            if (chart.cache.thresholdCustomPoints.hasOwnProperty(key)) {
                var type = key,
                    root1 = root;
                if (thresholdEventObj.grid) {
                    root1 = flatten.get(`pane${thresholdEventObj.grid}`);
                }
                var bubblePoints = root1.get(type);
                if (bubblePoints) {
                    root1.set(type, root1.concat(thresholdEventObj));
                } else {
                    root1.set(type, thresholdEventObj);
                }
            }
        }
    }

    // for other custom points
    let { customPoints } = chart.cache;
    if (customPoints) {
        for (let customPointsName in customPoints) {
            if (customPoints.hasOwnProperty(customPointsName)) {
                each(customPoints[customPointsName], (points, key) => {
                    let [seriesIndex, subSeriesIndex] = key.split('_').map(Number),
                        renderer = chart.renderer[seriesIndex];

                    if (isFunction(renderer)) {
                        // the _data is needed for our search model
                        points.forEach((point, itemIndex) => {
                            point.seriesIndex = seriesIndex;
                            point.subSeriesIndex = subSeriesIndex;
                            point._key = DataProcessor.helpers.generatePointKey(customPointsName, seriesIndex, subSeriesIndex, itemIndex);
                            point._data = dataHash(chart, renderer, point);
                        });
                    }
                });
                root.set(customPointsName, customPoints[customPointsName]);
            }
        }
    }

    if (this.pushFlattenAnnotation) {
        this.pushFlattenAnnotation();
    }

    /*
    ANNOTATION COMPONENT POINTS
     */
    if (chart.components) {
        chart.components.forEach(component => {
            let eventPoints = component.events;
            if (root.get(eventPoints.type)) {
                root.get(eventPoints.type)[eventPoints.id] = eventPoints.config[eventPoints.id];
            } else {
                root.set(eventPoints.type, eventPoints.config);
            }
        });
    }

    function getJoinbyKey(type, mode) {
        if (mode == 'charttype') {
            return type;
        } else {
            return getSearchMode(type, tooltipShared || processor.noofBarEnabledSeries);
        }
    }
};

DataProcessor.prototype.getFlattenedSeriesData = function(point) {
    let processor = this,
        chart = processor.chartObj,
        joinbykey = chart.chartTypes[point.seriesIndex].name,
        flatten = chart.dataObject.flattenPoints.get(joinbykey)[`${point.seriesIndex}_${point.subSeriesIndex}`];
    return flatten;
};

DataProcessor.prototype.removeFlattenData = function(point) {
    let flatten = this.getFlattenedSeriesData(point),
        position = point.itemIndex;

    flatten.splice(position, 1);
    for (let i = position; i < flatten.length - 1; i++) {
        flatten[i].itemIndex = position++;
    }
};

DataProcessor.prototype.pushFlattenAnnotation = function() {
    let processor = this,
        chart = processor.chartObj,
        chartdata = chart.seriesdata,
        annotationFlatten = (chart.annotation.flatten = []),
        root = this.flattenPoints.get('pane0_0'); // TODO

    annotationFlatten.push(function() {
        let annotation = chart.annotation.events; // TODO: rename
        chart.cache.parser = {}; // TODO: enhance

        // Flushing previous points
        annotation.forEach(a => {
            let _ = a.parser.event;
            root.set(_.type, null);
        });

        annotation.forEach(a => {
            let _ = a.parser.event;
            if (!root.get(_.type)) {
                root.set(_.type, {});
            }

            root.get(_.type).superKey = _.superKey;
            root.get(_.type).size = _.size;
            let seriesIndex = a.data.seriesIndex;
            let id = _.id; // + (defined(seriesIndex) ? '_' + seriesIndex : '');

            let arr = (root.get(_.type)[id] = root.get(_.type)[id] || []);
            arr.disabled = () => (defined(seriesIndex) ? chartdata[seriesIndex].disabled : false);
            arr.push(invokeFunction(_.data, a.data, chart, a.xy, a)); // TODO: enhance

            // let id = pick(_.id, 'zc-common');
            let _key = `${_.type}_${id}_${a.data._key}`;
            // chart.cache.parser[a.data._key] = chart.cache.parser[a.data._key] || {};
            chart.cache.parser[a.data._key] = a; // TODO: work
            chart.cache.parser[a.data._key]._highlighter = invokeFunction(
                isObjectPropDefined(chart.cache, `parser.${[a.data._key]}.parser.event.highlighter`, true),
                chart
            );
        });
    });
};

function dataHash(chart, renderer, point, y_axisorder, stacked) {
    var highlighter = renderer.highlighter;
    return highlighter.data(point);
}

function eventDisabled(series, type, chart) {
    /*
     * returns true if the series is filtered or event is disabled for that series
     */
    var plotOptions = series.plotoptions;
    var commonPlotOptions = chart.systemConf.chart.plot.plotoptions[type];
    return series.disabled || validateNullVal(getEventConf(plotOptions, commonPlotOptions, chart, 'enabled'), true) == false;
}

function generateCustomData(data, seriesIndex) {
    return {
        seriesIndex: seriesIndex,
        data: [data]
    };
}

function checkSegmentSelection(chart, renderer, defaultChartType) {
    var rp = renderer.commonRendererProp().params,
        plotOptions = rp.commonPlotOptions,
        markerOptions = rp.markerOptions || {},
        symbol = markerOptions.symbol,
        // check sankey link type
        link = plotOptions.link || {};

    if (symbol == 'line' || symbol == 'vertical-line') {
        return 'segment';
    } else if (link.curve) {
        return `${defaultChartType}Link${toCapitalize(link.curve)}`;
    } else {
        return defaultChartType;
    }
}

function getDataToFlatten(customdata, chart) {
    var gridlist = customdata ? [{ data: customdata }] : [{ data: chart.dataObject.processedseriesData }];
    if (chart.dataObject.pivot) {
        if (customdata) {
            gridlist = [
                {
                    row: customdata[0].row,
                    column: customdata[0].column,
                    data: customdata,
                    index: customdata[0].index
                }
            ];
        } else {
            var pivotLayout = chart.dataObject.hierarchical.pivotLayout,
                layers = pivotLayout.getRootNodeList('layer'),
                panes = layers.panes;
            gridlist = objectValues(panes);
        }
    }
    return gridlist;
}

function flatHierarchyData(data, arr) {
    if (!data) {
        return;
    }

    if (!isArray(data[0])) {
        arr.push(data);
        return;
    }
    data.forEach(d => {
        flatHierarchyData(d, arr);
    });
}

export function defaultHash(d, chart, series, item, type, _key) {
    d.seriesIndex = series || 0;
    d.subSeriesIndex = 0;
    d.itemIndex = item || 0;
    d.sortedItemIndex = item || 0;
    d.rendererIndex = 0;
    // d.componentIndex = 0;
    d.yaxisOrder = 0;
    d.paneid = '0_0';
    d._key = DataProcessor.helpers.generatePointKey(chart.id, type, series, 0, item);
    d.__key = _key;
    d._xid = chart.dataset.getActualX(d);
    d.type = type;
    return d;
}

Registry.setComponent('events.dataHash', defaultHash);
