// @zohocharts/zc/src/core/data-processor/minMaxYValues.js
/**
 *
 * @param index
 * @param correctionFactor
 * @returns {*}
 */
//minMaxYValues.js
import { DataProcessor } from '../data-processor';
import { pick, getPropVal, isObjectPropDefined, defined, arrayIncludes, isUndefined, objectValues } from '../../utils/zcutils';
import { chartcategory, charttype } from '../main-zc-init';
import { d3_extent, d3_merge } from '../d3-constants';
import { TIME_DATATYPE, NUMERIC_DATATYPE } from '../datatype-constants';
import { isBarChartType } from '../../events/events-utils';
import { getThresholdValues } from '../../threshold/threshold-utils';
import { LOG } from '../scale-type-constant';
import { getAxisObject, getAxisOrient } from '../../axis/axisutils';
// import { numericAxisParser } from '../../scale-parser/axis/numeric-axis-parser';
import { expressionEvalRangeForLinear } from '../../scale/evaluators/evalRangeNumericAxisParser';
import { expressionEvalDomainForLinear } from '../../scale/evaluators/evalDomainNumericAxisParser';
import { Scale } from '../../scale/scale-component';
import { BOTTOM_ALIGN, TOP_ALIGN } from '../align-constants';
import { extendScaleParser } from '../../parser/parser-utils';
import { Registry } from '../zc-registry';
// import { getMethodFromInstance } from "../../utils/instance-utils";

DataProcessor.prototype.getMinMaxY = function(index = 0, correctionFactor, excludeDatalabels, applyZoom = true) {
    var systemconf = this.chartObj.systemConf,
        columnInfo = this.dataset.getColumn('y', index),
        rangeOperators = {};
    var orient = getAxisOrient('y', this.chartObj.axes.rotated);
    rangeOperators.plotBound = {
        dimension: orient == BOTTOM_ALIGN || orient == TOP_ALIGN ? 'width' : 'height',
        customPlotBound: this.chartObj.systemConf.chart,
        plotSize: this.chartObj.plotarea,
        defaultPlotSize: this.chartObj.plotarea,
        color: this.chartObj.imagePallete || this.chartObj.seriesColor,
        boundExtreme: 0
    };
    const numericAxisParser = Registry.getComponent('axis.parser.numeric');
    var domainOperators = numericAxisParser('y', index, this.chartObj, correctionFactor, excludeDatalabels, applyZoom, true); //needOnlyExtremes apply for time datatype Y axis
    var parser = {
        type: 'numeric',
        field: columnInfo,
        scaleType: columnInfo.scaleType,
        reverse: pick(this.yaxesData[index], {}).reversed,
        data: [],
        domain: {
            operators: domainOperators,
            evaluator: expressionEvalDomainForLinear
        },
        range: {
            operators: rangeOperators,
            evaluator: expressionEvalRangeForLinear
        }
    };
    // if(this.chartObj.viewFlow && this.chartObj.viewFlow.scales[columnInfo.id]){
    //     extendScaleParser(parser, this.chartObj.viewFlow.scales[columnInfo.id].conf);
    // }
    
    if (this.chartObj.viewFlow) {
        let _scale = this.chartObj.viewFlow.getScaleByColumnName(columnInfo, 'y');
        if(_scale){
            extendScaleParser(parser, _scale.conf, this.chartObj);
        }
        
    }
    // if (systemconf.viewParser) {
    //     if (systemconf.viewParser.scale && systemconf.viewParser.scale[columnInfo.name]) {
    //         extendScaleParser(parser, systemconf.viewParser.scale[columnInfo.name]);
    //     }
    // }
    var newScale = new Scale('y', index, parser, this.chartObj);
    var domain = newScale._parseDomain();
    if (!this.yRange[index]) {
        this.yRange[index] = newScale.config.tempDomain;
    }
    this.hasNegativeValue[index] = newScale.config.hasNegativeValue && newScale.config.hasNegativeValue[index];
    return domain;

    /*
    SCALE PARSER
     */
    var processor = this,
        yAxisConfig = processor.yaxesData,
        hasNegativeValue = processor.hasNegativeValue,
        baseline = processor.baseline,
        yRange = processor.yRange,
        chart = processor.chartObj,
        dataSet = processor.dataset,
        stacked = processor.isstacked() || processor.isPercentChart,
        yAxisData = pick(yAxisConfig[index], {}),
        dataType = dataSet.getDataType('y', index),
        columnInfo = dataSet.getColumn('y', index),
        minY = yAxisData.minRange,
        maxY = yAxisData.maxRange,
        correction = 0,
        yEncode = processor.getEncode('y', index),
        axisObj = getAxisObject(chart, 'y', index);

    if (!yEncode) {
        return ['NoData', 'NoData'];
    }

    if (columnInfo.scaleType === LOG) {
        correctionFactor = 0;
        if (defined(axisObj.logDomain)) {
            // save baseline
            processor.saveBaseLine(index, axisObj.logDomain, minY);
            return axisObj.logDomain;
        }
    }

    // var filterFactors = ["zoom"],
    //     rawFactors = ["userdata", "threshold", "levelMarker"],
    //     correctionFactors = ["bubble", "marker", "datalabels"];
    //
    // // Filters
    // filterFactors
    // .some(factor => {
    //     var filteredExtent = getMethodFromInstance(processor, factor)("y", index);
    //
    //     if (filteredExtent) {
    //         return filteredExtent;
    //     }
    // });
    //
    // // Raw data
    // var cache;
    // if (!cache) {
    //     var rawExtent = rawFactors
    //         .map(factor => {
    //             var extent = getMethodFromInstance(processor, factor)("y", index) || [];
    //             return extent;
    //         });
    //
    //     cache = d3_extent(d3_merge(rawExtent));
    // }
    //
    // //Corrections
    // var finalExtent = [];
    // correctionFactors
    // .forEach(factor => {
    //     var correction = getMethodFromInstance(processor, factor)("y", index) || 0;
    //     finalExtent[0] -= correction / 2;
    //     finalExtent[1] += correction / 2;
    // });

    // [**MIGRATION**] zoom --> brush
    // #ZC2016 $index
    if (isObjectPropDefined(chart.systemConf, `chart.brush.y[${index}]`) && applyZoom) {
        var zoomedY = chart.systemConf.chart.brush.y[index],
            zoomedExtent = [zoomedY.minRange, zoomedY.maxRange];

        processor.saveBaseLine(index, zoomedExtent, null, true);
        return zoomedExtent;
    }

    correctionFactor = pick(correctionFactor, chart.datalabels.hasFixedPosition ? 0.00000001 : 0.05);
    baseline[index] = getPropVal(yAxisData, 'baseline.show') === true ? yAxisData.baseline.value : null;

    if (!yRange[index]) {
        // save the min and max of y at yRange
        var temp;

        if (processor.isWaterfallEnabled) {
            temp = processor.processDataForWaterFall(index);
        } else if (stacked) {
            temp = _updateMinMaxforStackedData(processor, index);
        } else {
            temp = updateminMax(index, processor);
        }

        if (temp) {
            var yAxis = pick(getPropVal(chart.userdata, `chart.axes.yaxis[${index}]`), {}),
                thresholdData = getThresholdValues(yAxis),
                levelMarkerData = DataProcessor.helpers.hasLevelMarkerData(processor, index) ? processor.overLevelMarker(index) : [];

            temp = temp.concat(thresholdData, levelMarkerData);
            temp = d3_extent(temp);
        }

        yRange[index] = temp;
    }

    var extentY = yRange[index];
    if (isUndefined(extentY)) {
        return ['NoData', 'NoData'];
    }

    processor.saveBaseLine(index, extentY, minY);
    var hasDefaultBase = dataType == NUMERIC_DATATYPE && _hasDefaultBase(yEncode, processor, chart),
        hasDefaultMin = hasDefaultBase && (hasNegativeValue[index] != null ? hasNegativeValue[index] > 0 : true);

    //If extentY = [undefined, undefined], don't change the extent with '0'
    if (hasDefaultBase && extentY[1] - extentY[0] === 0) {
        extentY = d3_extent(d3_merge([extentY, [0]])); // Test_511.json & Test_512.json
    }

    if (processor.isBubbleEnabled && dataType !== TIME_DATATYPE && isUndefined(yAxisData.minRange) && isUndefined(yAxisData.maxRange)) {
        //if 'hasDefaultMin' is present, apply that. so that we can adjust the bubbles within range.
        extentY[0] = hasDefaultMin ? 0 : extentY[0];

        temp = processor.overBubbles(extentY, 'y', index);
        extentY = [extentY[0] - temp[0], extentY[1] + temp[1]];
    }

    if (DataProcessor.helpers.hasMarkerCorrection() && dataType !== TIME_DATATYPE) {
        temp = processor.overMarker(extentY, 'y', index, correctionFactor);
        extentY = [hasDefaultMin ? 0 : extentY[0] - temp[0], extentY[1] + temp[1]];
    }

    var diff = extentY[1] - extentY[0];

    if (diff === 0) {
        diff = 1;
    }

    if (!(processor.isPercentChart || processor.hasLevelMarker || processor.isWaterfallEnabled)) {
        correction = diff * correctionFactor;
    }

    if (isUndefined(minY)) {
        minY = hasDefaultMin ? 0 : extentY[0] - correction;
    }

    if (isUndefined(maxY)) {
        maxY = extentY[1] + correction;
    }

    var correctionOverDatalabels = 0;

    // no need to find correction datalabel,if correction factor is 0
    if (dataType !== 'time' && !excludeDatalabels && (correctionFactor !== 0 || columnInfo.scaleType === LOG)) {
        var ignoreDatalabels = chart.datalabels.hasFixedPosition && defined(yAxisData.maxRange) && yAxisData.maxRange <= extentY[1];
        correctionOverDatalabels = ignoreDatalabels ? 0 : processor.getCorrectionOverDataLabels([minY, maxY], 'y', index);
    }

    if (!hasDefaultMin && !chart.datalabels.hasFixedPosition) {
        minY = minY - correctionOverDatalabels || 0;
    }

    maxY = maxY + correctionOverDatalabels || 0;

    if (processor.isWaterfallEnabled) {
        var domain = processor.getCorrectionForWaterfall([minY, maxY], 'y', index, chart, correctionOverDatalabels);
        minY = domain[0];
        maxY = domain[1];
    }

    return [minY, maxY];
};

function updateminMax(index, dataObj) {
    // var yValues = categoriesStore["y"+index];
    var yValues = d3_merge(dataObj.seriesExtermes[index]);
    var minmaxY = yValues.length != 0 ? d3_extent(yValues) : yValues.enabled ? [0, 0] : null; // yvalues.enabled for referring that axis is visible or not.If the axis is visible,but all the data contains 0 or null,we dont want to remove axis
    return minmaxY;
}

//PENDING: Stacked area missing subseries
function _updateMinMaxforStackedData(context, index) {
    var dataset = context.dataset,
        processedseriesData = context.processedseriesData,
        helpers = DataProcessor.helpers;

    var yEncode = context.getEncode('y', index),
        seriesData = yEncode.seriesData,
        seriesIndices = yEncode.seriesIndices;

    if (context.isPercentChart) {
        return [0, 100];
    }
    var extents;

    if (context.pivot) {
        var pivotLayout = context.hierarchical.pivotLayout,
            nodeList = pivotLayout.getRootNodeList('row'),
            layers = pivotLayout.layers,
            hasMultiCluster = nodeList.hasMultiCluster();
        seriesData = objectValues(layers.panes);
    }
    extents = seriesData.map((series, i) => {
        var seriesIndex = context.pivot ? i : seriesIndices[i],
            seriesAssist = context.seriesAssist[seriesIndex];

        if (arrayIncludes(seriesAssist.layout || [], 'stack')) {
            var stackIndex = context.pivot ? `${series.row}_${series.column}` : seriesAssist.stackIndex,
                layout = context.encodes.stack[stackIndex].layout;

            return layout.getExtent();
        } else {
            var data = d3_merge(series.data), //TODO: merge is not needed if subseries support is removed
                yaxiscolumnorder = helpers.getParsedYAxisOrder(processedseriesData[seriesIndex].yaxiscolumnorder),
                cType = context.allChartTypes[seriesIndex],
                yAccessor = d => {
                    var yVal = dataset.getY(d, yaxiscolumnorder);
                    return yVal;
                };

            if (chartcategory.rangePlot.has(cType)) {
                var mergedYranges = d3_merge(data.map(yAccessor));
                return d3_extent(mergedYranges);
            } else {
                return d3_extent(data, yAccessor);
            }
        }
    });

    return d3_extent(d3_merge(extents));
}

/**
 * Bar, Area chart has default base as 0
 * @param       {Object}  processor [description]
 * @param       {Object}  chartObj  [description]
 * @return      {Boolean}           [description]
 */
export function _hasDefaultBase(yEncode, processor, chartObj) {
    //handle multiY combinational charts level marker.
    var ismultiY = chartObj.getAxesCount('y') > 1;

    if (ismultiY) {
        var allChartTypes = processor.allChartTypes,
            seriesIndices = yEncode.seriesIndices,
            chartTypes_splitAxes = seriesIndices.map(d => allChartTypes[d]),
            check = d => (isBarChartType(d) && !chartcategory.rangePlot.has(d)) || d === charttype.area || d === charttype.stkdarea;

        return chartTypes_splitAxes.some(check);
    }

    return (
        ((processor.noofBarEnabledSeries && !processor.noofBoxplotEnabledSeries && !processor.noofBarRangeEnabledSeries) ||
        processor.noofAreaEnabledSeries) && (!chartObj.customVisual)
    ); //TODO: remove this variables
}

DataProcessor.prototype.ZC_BAR_getRangeY = function(index) {
    var yRange = this.yRange[index] || [];
    return d3_extent(d3_merge([yRange, [0]]));
};
