// @zohocharts/zc/src/core/data-processor/pre-data-processor.js
//pre-data-processor.js

/**
 *
 * @param chart
 * @constructor
 */
import { DataProcessor } from '../data-processor';
import { Registry } from '../zc-registry';
import { chartcategory } from '../main-zc-init';
import { _requireXDataSorting } from '../zcconf';
import { NUMERIC_DATATYPE, TIME_DATATYPE } from '../datatype-constants';
import { defined, isArray, arrayIncludes, splat, pick, isObjectPropDefined } from '../../utils/zcutils';
import { isPointChartType, isScatterChartType, isBarChartType } from '../../events/events-utils';
import { isAlluvialChartType, isAreaChartType } from '../charttype-helpers';
import { getGridList } from './pivot-layout-helpers';
import { SANKEY_CHARTNAME } from '../chartname-constants';

export function PreProcessor(chart) {
    this.chart = chart;
    this.data = chart.systemConf;
    this.initialize();
}

PreProcessor.prototype.initialize = function() {
    var processor = this,
        chart = processor.chart,
        data = processor.data,
        dataSplicer = chart.dataSplicer,
        dataset = chart.dataset,
        metadata_axes = dataSplicer.getEncode(),
        metadata_columns = dataSplicer.getColumns(),
        xColumnIndex = isAlluvialChartType(chart)? dataset.getColumnIdx('level', -1) : metadata_axes.x,
        xDataType = metadata_columns[xColumnIndex].datatype,
        dataProcessorHelpers = DataProcessor.helpers,
        isDiscrete = chartcategory.discrete.some(function(d) {
            return chart.chartTypes.indexOf(d) !== -1;
        }),
        cType = dataProcessorHelpers.getChartType(null, null, 0, chart.systemConf) || {},
        chartValue = cType.value || 9,
        chartName = cType.name || 'bar',
        facetHierarchy = data.facets && data.facets.hierarchy;

    // check if the data has hierarchy
    var isAllowedHierarchyCharts =
        isBarChartType(chartValue) ||
        isAreaChartType(chartValue) ||
        isScatterChartType(chartValue) ||
        isPointChartType(chartValue) ||
        chartName === 'line' ||
        (data.facets &&
            ['pie', 'funnel', 'pyramid', 'dial', 'web', 'packedbubble', 'treemap', 'sunburst', 'wordcloud'].indexOf(chartName) > -1);

    if ((this.isHierarchyLayout() && isAllowedHierarchyCharts) || facetHierarchy) {
        processor.hierarchical = new DataProcessor.HierarchicalOrdinalDataAdapter(this.chart);
        processor.isLayerHierarchy = !processor.hierarchical.pivot;
        processor.isPivot = processor.hierarchical.pivot;
        processor.isAxisCategoricalPivot = true;
        // update the chart data
        processor.chart.seriesdata = this.hierarchical.getSeries();
        if (this.hierarchical.pivot) {
            chart.chartTypes = DataProcessor.helpers.getAllChartTypesFromSeriesdata(
                processor.chart.userdata,
                processor.chart.seriesdata,
                'object'
            );
        }
    }
    var sytemConfOptions = chart.systemConf.chart.plot.plotoptions,
        plotoptions = sytemConfOptions && sytemConfOptions[chartName];
    chart.isAxisedHierarchyChart = plotoptions && plotoptions.axis && plotoptions.axis.enabled;
    chart.isLevelAxisCategory = isAlluvialChartType(chart);
    if (processor.hierarchical && processor.hierarchical.facetRendering) {
        let gridList = getGridList(processor);
        gridList.forEach(grid => {
            let gridDataLikeSystemConfig = {
                seriesdata: { chartdata: grid.data.map(_ => _.node), type: chart.systemConf.seriesdata.type }
            };
            if (isNestedHierarchy(chart, gridDataLikeSystemConfig)) {
                processor.isHierarchy = true;
                let dataAdapterObj = new (Registry.getComponent('hierarchyDataAdapter'))(grid.data, processor);
                grid.data = dataAdapterObj.getData();
            }
        });
    } else if ((isNestedHierarchy(chart) && !chart.isAxisedHierarchyChart) || (chart.customVisual && chart.viewFlow.hierarchy)) {
        processor.isHierarchy = true;
        let dataAdapterObj = new (Registry.getComponent('hierarchyDataAdapter'))(chart.seriesdata, processor);
        chart.seriesdata = dataAdapterObj.getData();
    }
    if(chart.isAxisedHierarchyChart){
        processor.isHierarchy = true;
    }


    // process the time data
    processor.processTimeData();

    // sort if defined
    if (_requireXDataSorting && (xDataType === NUMERIC_DATATYPE || (xDataType === TIME_DATATYPE && !isDiscrete))) {
        processor.sortData(xColumnIndex, xDataType);
    }

    const isHierarchicalSankey = chartName === SANKEY_CHARTNAME && plotoptions.hierarchical;
    if (isHierarchicalSankey) {
        const SankeyHierarchyLayout = Registry.getComponent('sankeyHierarchyLayout');   
        const sankeyHierarchyLayout = chart.dataObject.sankeyHierarchyLayout = new SankeyHierarchyLayout(chart, plotoptions.hierarchical);
        sankeyHierarchyLayout.initialize();
        chart.seriesdata = sankeyHierarchyLayout.dataAdapter.getData();
        this.hasHierarchyData = true;
    }
};

/**
 * Checks if the series has hierarchy
 * @return {boolean}
 */

PreProcessor.prototype.isHierarchyLayout = function() {
    var processor = this,
        chart = processor.chart,
        series = chart.systemConf.seriesdata.chartdata;

    return series.some(PreProcessor.helpers.hasInnerSeries);
};

/**
 *
 */

PreProcessor.prototype.processTimeData = function() {
    var processor = this,
        chart = processor.chart,
        timeDataInfo,
        timeDataIndices,
        timeDataInputFormat,
        hasTimeInputFormat,
        series = chart.seriesdata;

    timeDataInfo = PreProcessor.helpers.time.getDataInfo(processor.data, null, chart);
    timeDataIndices = timeDataInfo.indices;
    timeDataInputFormat = timeDataInfo.formatConf;

    hasTimeInputFormat = timeDataInputFormat.some(function(d) {
        return defined(d) && defined(d.inputformat);
    });

    if (timeDataIndices && timeDataIndices.length && hasTimeInputFormat) {
        let func = isNestedHierarchy(chart) ? loopHierarchy : loopFlatten;
        func(series, d => PreProcessor.helpers.time.applyFormats(d, timeDataInfo));
    }
};

/**
 *
 * @param xColPos
 * @return {*}
 */

PreProcessor.prototype.sortData = function(xColPos) {
    var processor = this,
        chart = processor.chart,
        data = processor.data,
        series = chart.seriesdata;

    series.forEach(function(currentSeries) {
        var currentData = currentSeries.data,
            dataIndex = data.metadata.columns[xColPos].dataindex;

        if (DataProcessor.helpers.is3D(currentData)) {
            if (isArray(currentData)) {
                currentData.forEach(function(subgroup) {
                    PreProcessor.helpers.linear.sort(subgroup, dataIndex);
                });
            }
        } else {
            PreProcessor.helpers.linear.sort(currentData, dataIndex);
        }
    });
    return data;
};

function isNestedHierarchy(chartObj, data) {
    data = pick(data, chartObj.systemConf);
    let chartType = DataProcessor.helpers.getChartType(null, NUMERIC_DATATYPE, 0, data);
    return arrayIncludes(chartcategory.hierarchy, chartType);
}

/**
 * Loop flat data
 * @param  {Array} data
 * @param  {Function} hook
 */
function loopFlatten(series, callbacks) {
    series.forEach(d => {
        let currentData = d.data;
        if (isArray(currentData)) {
            var is3DArray = DataProcessor.helpers.is3D(currentData);
            currentData.forEach(function(subgroup, j) {
                if (is3DArray && isArray(subgroup)) {
                    subgroup.forEach(dataPoint => executeCallBacks(callbacks, dataPoint));
                }
                if (!is3DArray) {
                    executeCallBacks(callbacks, subgroup);
                }
            });
        }
    });
}

/**
 * Loop hierarchy data
 * @param  {Array} data
 * @param  {Function} hook
 */
function loopHierarchy(parent, callbacks) {
    parent.forEach(d => {
        if (isArray(d[0])) {
            loopHierarchy(d, callbacks);
        } else {
            executeCallBacks(callbacks, d);
        }
    });
}

function executeCallBacks(callbacks, d) {
    callbacks = splat(callbacks);
    callbacks.forEach(callback => callback(d));
}
