// @zohocharts/zc/src/core/data-processor/sankey-processor.js
import { DataProcessor } from '../data-processor';
import { d3_ascending, d3_map, d3_set, d3_sum } from '../d3-constants';
import { defined, getPropVal, invokeFunction, isArray, simpleClone } from '../../utils/zcutils';
import { getSankeyLayout } from '../../sankey/sankey-chart-layout';
import { mathAbs, mathMax, UNDEFINED } from '../zc-constants';
import { getFormattedValue } from '../../formatter/formatter';
import { Registry } from '../zc-registry';

const SankeyLayout = Registry.getComponent('sankeyLayout');

const _sankeyProcessor = (DataProcessor.prototype.processSankeyData = function() {
    let self = this,
        { processedseriesData: processedSeries, dataset: dataSet } = self,
        xIndex = self.isLevelAxisCategory ? dataSet.levelidx(-1) : self.dataset.xidx,
        yIndex = self.isLevelAxisCategory ? dataSet.levelidx(-1) : self.dataset.yidx(0),
        zIndex = self.dataset.zidx,
        clridx = self.dataset.clridx,
        chartInstance = self.chartObj,
        cache = chartInstance.cache,
        sankeyOptions = getPropVal(chartInstance, 'systemConf.chart.plot.plotoptions.sankey') || {},
        updateLegendLayout = chartInstance.phase !== 'legendFilter',
        disabledItems = (cache.disabledItems = d3_set()),
        sankeyData = (cache.sankey = cache.sankey || {
            legend: [],
            legendItemsMap: d3_map()
        }),
        { legend, legendItemsMap } = sankeyData,
        layout;

    const shl = chartInstance.dataObject.sankeyHierarchyLayout;

    if (legend.length) {
        // create a map for faster access
        // TODO: Optimize code in legend-events filterSeriesNUpdate method
        legend.forEach(item => {
            if (item.disabled) {
                disabledItems.add(item[xIndex]);
            }
        });

        // processedSeries = shl.dependencies.convertSankeyData();
    }

    if (processedSeries != null && isArray(processedSeries)) {
        processedSeries.forEach((series, seriesIndex) => {
            // first add the chart type
            var { name, type } = DataProcessor.helpers.getChartType(null, null, seriesIndex, self.data),
                renderEmptyLink = getPropVal(self.data, `chart.plot.plotoptions.${name}.link.renderEmpty`),
                negativeAsAbsolute = getPropVal(self.data, `chart.plot.plotoptions.${name}.showAsNegative`) === 'absolute',
                sankeyLinkOptions = getPropVal(self.data, `chart.plot.plotoptions.${name}.link`),
                showExitLink = sankeyLinkOptions.exit.enabled;

            dataSet.seriesTypes.add(type);

            // convert the data to 3d if it is not
            if (isArray(series.data) && !DataProcessor.helpers.is3D(series.data)) {
                series.data = [series.data];
            }

            // currently for formatting, only x value is supported
            let format = (value, type) => {
                var column = dataSet.getColumn(type, 0);
                return getFormattedValue(column[column.datatype], value, column.datatype, chartInstance);
            };

            series.data.forEach((subSeries, subSeriesIndex) => {
                
                let sourceAccessor = (point, original) => original ? dataSet.getX(point, 0) : format(dataSet.getX(point, 0), 'x'),
                    targetAccessor = (point, original) => original ? dataSet.getY(point, 0) : format(dataSet.getY(point, 0), 'y'),
                    exitAccessor = point => {
                        if (showExitLink) {
                            return true;
                        } else {
                            return format(dataSet.getY(point, 0), 'y');
                        }
                    },
                    weightAccessor = point => {
                        let z = dataSet.getZ(point);
                        if (z < 0) {
                            return negativeAsAbsolute ? mathAbs(z) : 0;
                        }
                        return z;
                    },
                    colorAccessor = () => undefined,
                    filterAccessor = () => false;

                if (shl) {
                    format = shl.dependencies.dataFormatter;
                    sourceAccessor = shl.accessor('source');
                    targetAccessor = shl.accessor('target');
                    exitAccessor = shl.accessor('exit');
                    weightAccessor = shl.accessor('weight');
                    colorAccessor = shl.accessor('color');
                    filterAccessor = shl.accessor('filter');
                }
                const accessors = {source: sourceAccessor, target: targetAccessor, weight: weightAccessor, 
                    exit: exitAccessor, color: colorAccessor, filter: filterAccessor};

                let savedZoomTransform =  self.sankeyLayout ? self.sankeyLayout.zoomTransform : chartInstance.sankeyZoomTransform ? chartInstance.sankeyZoomTransform : null;  
                layout = self.sankeyLayout = new SankeyLayout( subSeries, accessors, disabledItems);
                layout.zoomTransform = savedZoomTransform;

                // override sankey layout methods
                shl ? Object.assign(layout, shl.layout('override')) : null;
                
                subSeries.forEach((point, itemIndex) => {
                    if (
                        SankeyLayout.isValidPoint(sourceAccessor(point), targetAccessor(point), exitAccessor(point), weightAccessor(point))
                    ) {
                        // need in sankey renderer for mapping
                        point._key = DataProcessor.helpers.generatePointKey(
                            chartInstance.id,
                            'pane0_0',
                            seriesIndex,
                            subSeriesIndex,
                            itemIndex
                        );
                        point._type = 'link';
                        if (!targetAccessor(point)) {
                            point.exit = true;
                        }
                    }
                });
            });
        });

        if (layout.nodeLength || layout.linkLength) {
            // create a layout calculate the weight of each node
            // we will create a seriesdata with this [xValue, yValue, zValue]
            // it will help with existing flow of legend and datalabel rendering
            let chartWidth, chartHeight;
            if (chartInstance.systemConf.chart.width) {
                chartWidth = invokeFunction(chartInstance.systemConf.chart.width, chartInstance);
            }
            if (chartInstance.systemConf.chart.height) {
                chartHeight = invokeFunction(chartInstance.systemConf.chart.height, chartInstance);
            }
            let dummyLayoutConstructor = getSankeyLayout({ node: { width: 1 }, link: {}, iterations: sankeyOptions.iterations /* TODO: change for better performance */ }, [
                    [0, 0],
                    [chartWidth || chartInstance.canvasarea.width, chartHeight || chartInstance.canvasarea.height]
                ], UNDEFINED, UNDEFINED, UNDEFINED, chartInstance),
                dummyLayout = dummyLayoutConstructor(layout.graph),
                dummyLayoutForLegend = (sankeyData.legendSankeyLayout = dummyLayoutConstructor(layout.completeGraph)),
                layerCompare = (a, b) => d3_ascending(a.layer, b.layer);
            self.sankeyLayout.layout = dummyLayout;

            if (shl) {
                layout.graph.links.forEach(link => link.point._root.link = link);   // link reference in point for easy access
            }
            // sor the nodes based on layer so that the legend renders in order

            dummyLayout.nodes.sort(layerCompare);
            dummyLayoutForLegend.nodes.sort(layerCompare);

            normalizeNodes(dummyLayoutForLegend.nodes, dummyLayout.nodes);
        }

        return sankeyData;
    }

    function normalizeNodes(nodes, pointNodes) {
        let nodesMapped = d3_map(pointNodes, n => n.name);

        if (updateLegendLayout) {
            legendItemsMap.clear();
            legend.length = 0;

            nodes.forEach((node, i) => {
                let pointNode = nodesMapped.get(node.name),
                    t0 = [];

                t0[xIndex] = t0[yIndex] = node.name;
                t0[zIndex] = node.value;
                clridx >= 0 && !defined(t0[clridx]) ? t0[clridx] = node.point[clridx] : null;
                t0._type = 'node';
                t0.itemIndex = i;
                t0.seriesIndex = 0;
                t0.disabled = disabledItems.has(node.name);
                t0._format = node._format;     // sankey - time format issue
                t0._value = node._format === 'y' ? node.point[yIndex] : node.point[xIndex];
                t0.level = node.level;
                // t0._yaxisOrder = mathMax(node._level.split('_').length - 1, 0);     // TODO: Update proper yaxisorder based on depth level
                t0.original = simpleClone(t0);   // add original point values in formatted point
                t0.original[xIndex] = t0.original[yIndex] = node.nameOriginal;
                t0._axisOrder = node._axisOrder;
                t0.arcColor = node.color;
                t0.node = node;
                t0.clickSelection = node.point.clickSelection;
                legend.push(t0);

                node.point = t0;
                node.index = i;

                if (pointNode) {
                    pointNode.index = i;
                    pointNode.point = t0;
                }

                legendItemsMap.set(node.name, { index: i, point: t0 });
            });
        } else {
            nodes.forEach(node => {
                let { point, index } = legendItemsMap.get(node.name),
                    pointNode = nodesMapped.get(node.name);

                node.index = index;
                node.point = point;

                if (pointNode) {
                    pointNode.index = index;
                    pointNode.point = point;
                }

                point.disabled = disabledItems.has(node.name);
            });
        }
    }
});

export { _sankeyProcessor };
