// @zohocharts/zc/src/core/data-processor/sankey-hierarchy-processor.js
import { getFormattedValue } from "../../formatter/formatter";
import { defined, getDepth, getDistinct, getLevel, getPropVal, hasSameParent, invokeFunction, isAncestorDisabled, isArray, isObject, isString, pick, setPropVal, simpleClone } from "../../utils/zcutils";
import { d3_hierarchy, d3_sum } from "../d3-constants";
import { DataProcessor } from "../data-processor";
import { mathAbs } from "../zc-constants";
import { Registry } from "../zc-registry";

Registry.setComponent('sankeyHierarchyLayout', SankeyHierarchyLayout)
/**
 * 
 * @param {*} chart - chartObj
 * @param {*} callback - callback support for customization
 */

export function SankeyHierarchyLayout(chart, callback) {
    const self = this,
        helpers = DataProcessor.helpers;

    const chartdata = chart.systemConf.seriesdata.chartdata;

    this.initialize = function (data = chartdata) {
        const HierarachyDataAdapter = Registry.getComponent('hierarchyDataAdapter');
        const dataAdapter = self.dataAdapter = new HierarachyDataAdapter(data, {});
        const seriesdata = dataAdapter.getData();


        sankey_helpers_collector.call(this, chart);    // collect sankey utils, dependencies and helpers
        invokeFunction(true, callback, self, data, chart);

        chart.dataObject.hasHierarchicalData = true;
        self.completeGraph = self.dependencies.convertSankeyData(seriesdata);
        self.completeGraph.total = self.dependencies.getTotalData(seriesdata);
    };

    return this;
}

function sankey_helpers_collector(chart) {
    const _this = this,
        helpers = DataProcessor.helpers;

    const dataset = chart.dataset,
        userdata = chart.userdata,
        chartdata = userdata.seriesdata.chartdata,
        dataSplicer = chart.dataSplicer,
        axes = dataSplicer.getEncode(),
        columns = dataSplicer.getColumns(),
        // yaxiscolumnorder = helpers.getParsedYAxisOrder(),
        // level_axes = axes.level,
        // y_axes = axes.y[yaxiscolumnorder],
        xidx = dataset.levelidx(-1),
        zidx = dataset.zidx,
        clridx = dataset.clridx,
        seriesdata = this.dataAdapter.getData(),
        sankeyOptions = getPropVal(chart.systemConf, 'chart.plot.plotoptions.sankey'),
        sankeyNodeOptions = getPropVal(sankeyOptions, 'node'),
        nodeFillColor = sankeyNodeOptions.fillColor,
        sankeyLinkOptions = getPropVal(sankeyOptions, 'link'),
        linkFillColor = sankeyLinkOptions.fillColor,
        showExitLink = sankeyLinkOptions.exit.enabled,
        negativeAsAbsolute = getPropVal(sankeyOptions, 'showAsNegative') === 'absolute';

    this.helpers = { sankey: {} };

    // add util methods here
    this.utils = {
        getLevel, getTopLevel, getSeriesIndex, isDisabled, getDepth,
        getColumn, getColumnIndexByLevel,
        getAxisFormatByLevel, getProcessedNode, getProcessedNodes, getTotalValue
    }

    // add dependency methods here
    this.dependencies = {
        dataFormatter: function (value, level) {
            const column = getColumnByLevel(level);
            return getFormattedValue(column[column.datatype], value, column.datatype, chart);
        },
        convertSankeyData: function (data = seriesdata) {
            // const sourceIndex = 0, targetIndex = 1, weightIndex = 2;

            let root = d3_hierarchy(data, _this.helpers.sankey.hierarchy.data)
                .sum(_this.helpers.sankey.hierarchy.sum);

            _this.maxHeight = root.height;   // for depth calculations

            // Convert hierarchical data to Sankey data
            let flattenData = root.descendants();
            let sankeyNodes = [],
                sankeyLinks = _this.helpers.sankey.hierarchy.links(flattenData);

            const completeGraph = { hierarchy: root, flatten: flattenData, links: sankeyLinks, nodes: sankeyNodes };
            return completeGraph;
        },
        updateSankeyData: function () {
            const root = _this.completeGraph.hierarchy;
            root.sum(_this.helpers.sankey.hierarchy.sum);

            //update link values binded to nodes
            root.each(function (node) {
                node.point ? node.point[zidx] = node.value : null;
                node.link ? node.link.point[zidx] = node.value : null;
            })
        },
        getTotalData: function (data) {
            let root = d3_hierarchy(data, _this.hierarchy('data'))
                .sum(_this.hierarchy('total').sum);
            return { hierarchy: root };
        }
    }

    // add hierarchy data convertor callback methods here
    this.helpers.sankey.hierarchy = {
        sum: function (d) {
            if (isAncestorDisabled(seriesdata, d)) {
                return 0;
            }
            let y = d[zidx] || 0;
            return y < 0 ? mathAbs(y) : y;
        },
        data: function (node) {
            if (isArray(node[0])) {
                return node;
            } else {
                return null;
            }
        },
        links: function (flattenData) {
            const sankeyLinks = [];
            flattenData.forEach((node, i) => {
                const point = [];
                if (node.depth < 2) { return };
                const d0 = node.parent.data,
                    d1 = node.data;

                //TODO: Try using nodes to get sankey links
                point[xidx] = [d0.seriesname, node.children ? d1.seriesname : d1[0]];    // [source, target];
                point[zidx] = node.value;
                clridx >= 0 ?  point[clridx] = node.clrvalue = node.data.clrvalue : null; //clr value
                point.level = d1.level;
                point.parent = node.parent.link;
                point.depth = node.depth;
                point.height = node.height;  //* Used in sankey-layout _format and _axisOrder calculation which is used in legend text
                point.color = _this.helpers.sankey.styles.fillColor(node, d1.level, d0.color);
                point.arcColor = point.color && point.color.link;
                point._root = node;
                node.point = point;

                sankeyLinks.push(point);
            })
            return sankeyLinks;
        }
    }

    //TODO:= override hierarchy function to handle color case
    if (defined(clridx)) {
        _this.hierarchy('sum', function (d) {
            if (isAncestorDisabled(seriesdata, d)) {
                d.clrvalue = 0;
                return 0;
            }
            
            d.clrvalue = d[0] && d[0].clrvalue ? d3_sum(d, d => d.clrvalue) : d[clridx] || 0

            let y = d[zidx] || 0;
            return y < 0 ? mathAbs(y) : y;
        })
    }

    // add sum method to get total sum values without disabling
    this.helpers.sankey.hierarchy.total = {
        sum: function (d) {
            let y = d[zidx] || 0;
            return y < 0 ? mathAbs(y) : y;
        }
    }


    // add formatter methods here
    this.helpers.sankey.formatter = {
        source: function (point) {
            return _this.dependencies.dataFormatter(point[xidx][0], getTopLevel(getLevel(point)));
        },
        target: function (point) {
            return _this.dependencies.dataFormatter(point[xidx][1], getLevel(point));
        },
        exit: function (point) {
            return _this.dependencies.dataFormatter(point[xidx][1], getLevel(point));
        }
    }

    // add accessor methods here
    this.helpers.sankey.accessor = {
        source: function (point, original) {
            if (original) { return point[xidx][0] };
            const level = getTopLevel(getLevel(point));
            const formattedValue = _this.formatter('source')(point);
            return defined(formattedValue) ? formattedValue + getDepth(level) : formattedValue;
        },
        target: function (point, original) {
            if (original) { return point[xidx][1] };
            const level = getLevel(point);
            const formattedValue = _this.formatter('target')(point); 
            return defined(formattedValue) ? formattedValue + getDepth(level) : formattedValue;
        },
        exit: function (point, original) {
            return showExitLink ? true : original ? point[xidx][1] : _this.formatter('exit')(point);
        },
        weight: (point, original) => {
            // if (point.disabled) { return 0 };
            const z = dataset.getZ(point);
            return z < 0 ? negativeAsAbsolute ? mathAbs(z) : 0 : z;
        },
        /**
         * 
         * @param {*} point 
         * @param {string} [name='source'] - source | target | link. @default
         * @returns {string} color
         */
        color: (point, name = 'source') => {
            return point.color && point.color[name] || getPropVal(point, `parent.color.${name}`);
        },
        /**
         * point should be filtered from nodeSet and linkSet
         * @param {*} point
         * @param {string} [name='source'] - source | target | link. @default
         * @returns {boolean} - filtering condition
         */
        filter: (point, name = 'source') => {
            return !(point[zidx] > 0 ? true : name !== 'link');
        }
    }

    this.helpers.sankey.styles = {
        fillColor: (node, level, _color) => {
            const colorIterator = (node, colors = []) => {
                if (node) {
                    if (colors.length) {
                        const index = +level.split('_').pop();
                        return colors[index % colors.length];
                    } else {
                        return node.data.color || colorIterator(node.parent);
                    }
                }
            };

            const childrenColors = getPropVal(node.parent, 'data.plotoptions.childrenColors') || (node.depth == _this.maxHeight ? sankeyOptions.childrenColors : null);
            const sourceColor = colorIterator(node.parent);
            const targetColor = childrenColors ? colorIterator(node.parent, childrenColors) : node.data.color;
            return { source: sourceColor, target: targetColor, link: sourceColor };
        }
    }

    this.helpers.sankey.filter = {
        disableItems: function (d) {
            let nodes = getProcessedNodes(d), filteredSeriesData;
            const setDisabled = (node, depth = 0) => {
                if (node.depth === 0) { return };
                filteredSeriesData = getProcessedNode(node.data, chart.systemConf.seriesdata.chartdata);
                node.data.disabled = filteredSeriesData.disabled = true;
                if (depth >= 0) {
                    node.children && node.children.forEach(_ => isObject(_) ? setDisabled(_, depth + 1) : null);
                }
                if (depth <= 0) {
                    // isObject(node.parent) ? setDisabled(node.parent, depth - 1) : null;
                }
            };

            nodes.forEach(node => {
                setDisabled(node);
            });
        },
        enableItems: function (d) {
            let nodes = getProcessedNodes(d), filteredSeriesData;
            const setEnabled = (node, depth = 0) => {
                if (node.depth === 0) { return };
                filteredSeriesData = getProcessedNode(node.data, chart.systemConf.seriesdata.chartdata);
                node.data.disabled = filteredSeriesData.disabled = false;
                if (depth >= 0) {
                    node.children && node.children.forEach(_ => isObject(_) ? setEnabled(_, depth + 1) : null);
                }
                if (depth <= 0) {
                    isObject(node.parent) ? setEnabled(node.parent, depth - 1) : null;
                }
            };

            nodes.forEach(node => {
                setEnabled(node);
            })
        }
    }

    // methods to override in sankey layout
    this.helpers.sankey.layout = {};
    this.helpers.sankey.layout.override = {
        getAllConnectedPoints: function (point, flatten, accessor = {}) {
            let layout = this, points;
            // add default accessors
            accessor.visibility = accessor.visibility || (link => link);

            if (point._type === 'link') {
                if (flatten) {
                    points = [..._this.layout('traversal').getDependentPoints(point.point, accessor)]
                } else {
                    points = _this.layout('traversal').getDependentPoints(point.point, accessor);
                }
            } else if (point._type === 'node') {
                if (flatten) {
                    points = [..._this.layout('traversal').getDependentPoints(point.point, accessor)];
                } else {
                    points = _this.layout('traversal').getDependentPoints(point.point, accessor);
                }
            }
            return points;
        },
        /**
         * 
         * @param {object} link - can be a sankey node or returns 
         * @returns {array[]{}} - array of nodes and links connected to the respective node or link
         */
        getConnectedNodes: function (link) {
            // return [link.source, link.target];
            const accessor = {
                visibility: (link) => { return link; }
            }
            return _this.layout('traversal').getDependentPoints(link, accessor);
        }
    }

    // add sankey traversal methods here
    this.helpers.sankey.layout.traversal = {
        getDependentPoints: function (point, accessor = {}) {
            const points = [], level = point.level, traversal = _this.layout('traversal'), renderer = chart.renderer[0];

            if (point._type === 'link') {
                var link = point._key ? renderer.getLinkByPoint(point) : point;
                points.push(link);
                traversal.lookupLeft(points, link.source, level, 0, accessor.visibility);
                traversal.lookupRight(points, link.target, level, 0, accessor.visibility);
            } else if (point._type === 'node') {
                var node = renderer.getSankeyNodeByPoint(point);
                traversal.lookupLeft(points, node, level, 0, accessor.visibility);
                traversal.lookupRight(points, node, level, 0, accessor.visibility);
            }

            return getDistinct(points, accessor.distinct);
        },
        lookupLeft: function (points, node, level, depth = 0, accessor) {
            const _this = this;
            accessor = accessor || (link => link && link.point._visible !== false);     // visibility accessor
            // var parentLinkLevel = points.length ? points[points.length - 1].level : '';

            points.push(node);

            // change the target link based on link & depth
            var targetLinks = depth === 0 || node.level ? node.targetLinks : [node.targetLinks[0]];

            if (targetLinks.length) {
                targetLinks.forEach(function (targetLink) {
                    if (accessor(targetLink, 'target')) {
                        if ((depth > 0 || points[0]._type == 'link') && node.level && !hasSameParent(level, targetLink.level, node.layer, -1)) {
                            return;
                        }
                        points.push(targetLink);
                        _this.lookupLeft(points, targetLink.source, targetLink.level, depth + 1, accessor);
                    }
                });
            }
        },
        lookupRight(points, node, level, depth = 0, accessor) {
            const _this = this;
            accessor = accessor || (link => link && link.point._visible !== false);     // visibility accessor

            points.push(node); // change the target link based on link & depth

            var sourceLinks = !node.sourceLinks ? []
                : depth === 0 || node.level ? node.sourceLinks : [node.sourceLinks[0]];

            if (sourceLinks.length) {
                sourceLinks.forEach(function (sourceLink) {
                    if (accessor(sourceLink, 'source')) {
                        if ((depth > 0 || points[0]._type == 'link') && node.level && !hasSameParent(sourceLink.level, level, points[0].layer)) {
                            return;
                        }
                        points.push(sourceLink);
                        _this.lookupRight(points, sourceLink.target, sourceLink.level, depth + 1, accessor);
                    }
                });
            }
        }
    }

    // *******************************************************************************************************************/

    // add methods with dependent variables here
    function getColumnByLevel(level) {
        return columns[getColumnIndexByLevel(level)];
    }

    function getColumnByData(d) {
        return getColumnByLevel(getLevel(d));
    }

    /**
     * Returns the column based on data or level
     * @param {Array | string} d - might be data array or level string
     * @returns {Object} - column object
     */
    function getColumn(d) {
        return isArray(d) ? getColumnByData(d) : getColumnByLevel(d.level)
    }

    function getColumnIndexByLevel(level) {
        return dataset.getColumnIdx('level', getDepth(level) - 1);
    }

    function getAxisFormatByLevel(level) {
        //TODO: Change this function for axes.level
        const depth = getDepth(level);
        const index = _this.maxHeight - depth + 1;
        return index > 0 ? 'y' : 'x';
    }

    function getTotalValue(d) {
        const level = isArray(d) ? getLevel(d) : d;
        return getProcessedNode(level).value;
    }

    /**
     * 
     * @param {Array | string} d - might be data array or level string
     * @returns {Object} node
     */
    function getProcessedNode(d, node = _this.completeGraph.hierarchy) {
        const levelArray = isArray(d) ? getLevel(d).split('_') : isString(d) ? d.split('_') : [];
        while (levelArray.length) {
            node = node.children ? node.children[levelArray.shift()] : node.data ? node.data[levelArray.shift()] : node[levelArray.shift()];
        }
        return node;
    }

    //! Warn: It is currently based on sankey node. Change it to process based on sankey data 
    function getProcessedNodes(d, node = _this.completeGraph.hierarchy) {
        const targetLinks = d.node ? getPropVal(d.node, 'targetLinks.length') ? d.node.targetLinks : d.node.sourceLinks : [];
        const links = targetLinks.length ? targetLinks.map(_ => _.point) : [d];
        const nodes = [];
        links.forEach(link => {
            nodes.push(getProcessedNode(link, node));
        })
        return nodes;
    }

}

SankeyHierarchyLayout.prototype.accessor = function (name, value) {
    return arguments.length > 1 ? (this.helpers.sankey.accessor[name] = value, this) : this.helpers.sankey.accessor[name];
}

SankeyHierarchyLayout.prototype.formatter = function (name, value) {
    return arguments.length > 1 ? (this.helpers.sankey.formatter[name] = value, this) : this.helpers.sankey.formatter[name];
}

SankeyHierarchyLayout.prototype.styles = function (name, value) {
    return arguments.length > 1 ? (setPropVal(this.helpers.sankey.styles, name, value), this) : getPropVal(this.helpers.sankey.styles, name);
}

SankeyHierarchyLayout.prototype.filter = function (name, value) {
    return arguments.length > 1 ? (this.helpers.sankey.filter[name] = value, this) : this.helpers.sankey.filter[name];
}

SankeyHierarchyLayout.prototype.hierarchy = function (name, value) {
    return arguments.length > 1 ? (this.helpers.sankey.hierarchy[name] = value, this) : this.helpers.sankey.hierarchy[name];
}

SankeyHierarchyLayout.prototype.layout = function (name, value) {
    return arguments.length > 1 ? (this.helpers.sankey.layout[name] = value, this) : this.helpers.sankey.layout[name];
}


// util methods 
function getTopLevel(level, step = 0) {
    while (step-- >= 0) {
        level = level.slice(0, level.lastIndexOf('_'));
    }
    return level;
}

function getSeriesIndex(d) {
    let level = getLevel(d);
    return level ? +level.split('_')[0] : null;
}

function isDisabled(data, d) {
    const seriesIndex = getSeriesIndex(d);
    return defined(seriesIndex) && data[seriesIndex].disabled;
}
