// @zohocharts/zc/src/core/data-processor/stack-helpers.js
//stack-helpers.js

import { DataProcessor } from '../data-processor';
import { NUMERIC_DATATYPE } from '../datatype-constants';
import { isBarChartType } from '../../events/events-utils';
import { UNDEFINED } from '../zc-constants';
import { arrayIncludes, each, objectValues } from '../../utils/zcutils';
import { d3_merge } from '../d3-constants';
import { stack } from '../main-zc-init';

var stack_helpers = stack.helpers;
/**
 * TODO: revamp
 * Non Bar charts: adds dummy data and order series
 * @param  {Array} data      [description]
 * @param  {Object} dataset   [description]
 * @param  {Object} processor [description]
 */
export function prepareStackData(data, dataset, processor) {
    var helpers = DataProcessor.helpers,
        isCatNullX = helpers.isCatNull(dataset, 'x'),
        xCategories,
        xCategoriesLength;

    data.forEach(function(series, seriesIndex) {
        var chartType = helpers.getChartType(null, NUMERIC_DATATYPE, seriesIndex, processor.data),
            isStkd = processor.isstacked(chartType) || processor.isPercentChart;

        if (isStkd && !isBarChartType(chartType)) {
            var xMappingFunction = d => dataset.getX(d);

            var columnList, splitLeavesColumn;
            if (processor.categoryHierarchy) {
                columnList = processor.hierarchical.pivotLayout.getRootNodeList('column');
                splitLeavesColumn = columnList.splitColumnLeaves;
                var uniformLeafScale = processor.chartObj.systemConf.chart.axes.xaxis.uniformLeafScale;
                xCategories = uniformLeafScale
                    ? columnList.getUniqueDomainForCluster(0)
                    : columnList.getDomainForClusterSet(0, series.column);
                if (columnList.count && !splitLeavesColumn) {
                    xMappingFunction = d => dataset.getActualX(d);
                }
            }

            if (!xCategories) {
                xCategories = processor.getDomain(processor.getCategories('x'), 'x', true); // Fixed #ZC1179
                if (processor.pivot) {
                    xCategories = columnList.getUniqueDomainForCluster(0);
                }
                xCategoriesLength = xCategories.length;
            }

            series.data.forEach(function(d) {
                var currentkeys = d.map(xMappingFunction);
                for (var idx = 0, len = xCategoriesLength; idx < len; idx++) {
                    var key = xCategories[idx];
                    var haveKey = currentkeys.indexOf(key);
                    //aaranging order of data as per x order;
                    if (haveKey >= 0 && haveKey != idx) {
                        d.splice(idx, 0, d[haveKey]); //copying , Moving unordererd data to correct position i
                        d.splice(haveKey + 1, 1); // removing the copied data from old position havekey
                        currentkeys = d.map(xMappingFunction); // Rearranging the currentkeys
                    }
                    if (typeof currentkeys[idx] != UNDEFINED && isCatNullX(currentkeys[idx])) {
                        d.splice(idx, 1);
                    }
                    if (haveKey < 0) {
                        var duplicate = [];
                        var order = helpers.getParsedYAxisOrder(series.yaxiscolumnorder);
                        duplicate.splice(dataset.xidx, 0, key);
                        duplicate.splice(dataset.yidx(order), 0, 0); //dummy data add as 0
                        duplicate.isDummy = true;
                        // to remove stacked area with dummy data added
                        duplicate.isPrevDummy = idx == 0 ? true : d[idx - 1].isDummy;
                        if (idx == len - 1) {
                            duplicate.isNextDummy = true;
                        } else if (idx != 0) {
                            d[idx - 1].isNextDummy = true;
                        }

                        d.splice(idx, 0, duplicate);
                        currentkeys = d.map(xMappingFunction); // Rearranging the currentkeys
                    } else {
                        d.isDummy = false;
                    }
                }

                // If the series has duplicate x values, just remove that. Test_586.json, Test_587.json
                if (xCategoriesLength != d.length) {
                    d.splice(xCategoriesLength, d.length - xCategoriesLength);
                }
            });
        }
    });
}

export function AddDataToStackLayout(data, dataset, processor) {
    var helpers = DataProcessor.helpers,
        hasSplitLeaves = processor.pivot ? processor.hierarchical.pivotLayout.columns.splitColumnLeaves : false;
    data.forEach((seriesData, index) => {
        var seriesIndex = processor.pivot ? seriesData.row + seriesData.column + index : index,
            seriesAssist = processor.seriesAssist[seriesIndex];

        if (!seriesData.disabled && arrayIncludes(seriesAssist.layout || [], 'stack')) {
            var stackIndex = processor.pivot ? `${seriesData.row}_${seriesData.column}` : seriesAssist.stackIndex,
                stackTransform = processor.getEncode('stack', stackIndex),
                layout = stackTransform.layout,
                curdata = seriesData.data,
                yaxiscolumnorder = helpers.getParsedYAxisOrder(seriesData.yaxiscolumnorder),
                data = d3_merge(curdata), //TODO: merge is not needed if subseries support is removed
                valueAccessor = d => dataset.getY(d, yaxiscolumnorder),
                keyAccessor = d => dataset.getX(d);

            //ZC1931, ZC1932
            var chartType = helpers.getChartType(null, NUMERIC_DATATYPE, seriesIndex, processor.data);
            if (isBarChartType(chartType)) {
                // stack bar chart
                // #ZC2315 @param getCategories(0, false)

                var xCategories = processor.getCategories('x', 0, false),
                    xDomainValues = hasSplitLeaves ? xCategories : processor.getDomain(xCategories, 'x', true); // Fixed #ZC1179
                data = processor.updateSeriesdatabyCategories(data, xDomainValues); //TODO: IMPORTANT updateSeriesdatabyCategories should be done on the data processor init only once. now every renderer has this function
            }

            layout.add(data, keyAccessor, valueAccessor);
        }
    });
    let stackSortAccessor = processor.isStackedSort();
    if (stackSortAccessor) {
        let stackEncodes = processor.getEncode('stack'),
            keys = objectValues(stackEncodes);

        let valueAccessor = d => dataset.getY(d, 0),
            keyAccessor = d => dataset.getX(d);
        keys.map(d => {
            let stackLayout = d.layout;
            stackLayout.sort(keyAccessor, valueAccessor, stackSortAccessor);
        });
    }
}

export function saveStackInfo(processor, curdata, i, stackIndex, chartType_num) {
    var groupedByStack = processor.groupedByStack,
        obj = {
            si: i,
            d: curdata
        };

    if (isBarChartType(chartType_num)) {
        if (!groupedByStack.has(stackIndex)) {
            groupedByStack.set(stackIndex, [obj]);
        } else {
            var lastData = groupedByStack.get(stackIndex);
            groupedByStack.set(stackIndex, lastData.concat([obj]));
        }
        groupedByStack.get(stackIndex).stackLabelIndex = obj.si;
    }
}

export function convertToStackPercent(stackedData, dataset) {
    each(stackedData, (datum, stackIndex) => {
        var seriesData = datum.seriesData,
            stackLayout = datum.layout;

        seriesData.map(series => {
            var data = d3_merge(series.data), //TODO: merge is not needed if subseries support is removed
                keyAccessor = d => dataset.getX(d);

            stackLayout.toPercent(data, keyAccessor);
        });
    });
}

export var removeDummyData = stack_helpers.removeDummyData = function removeDummyData (series) {
    var removalElePos = new Array(0);
    series.data[0].map((d, i) => {
        if (d.isDummy) {
            removalElePos.push(i);
        }
        d.isDummy = false;
        d.isPrevDummy = false;
        d.isNextDummy = false;
    });

    for (var k = removalElePos.length - 1; k >= 0; k--) {
        series.data[0].splice(removalElePos[k], 1);
    }
    return series;
}

DataProcessor.prototype.getStackingIndices = function(index) {
    var context = this;

    var yEncode = context.getEncode('y', index),
        seriesData = yEncode.seriesData,
        seriesIndices = yEncode.seriesIndices;

    var stackIndices = new Array(0),
        nonStackseriesIndices = new Array(0),
        _stackData = new Array(0),
        _nonStackData = new Array(0);
    if (context.pivot) {
        var pivotLayout = context.hierarchical.pivotLayout,
            layers = pivotLayout.layers;
        seriesData = objectValues(layers.panes);
    }
    seriesData.map((series, i) => {
        var seriesIndex = context.pivot ? i : seriesIndices[i],
            seriesAssist = context.seriesAssist[seriesIndex];

        if (arrayIncludes(seriesAssist.layout || [], 'stack')) {
            var stackIndex = context.pivot ? `${series.row}_${series.column}` : seriesAssist.stackIndex;
            stackIndices.push(stackIndex);
            _stackData.push(series);
        } else {
            nonStackseriesIndices.push(seriesIndex);
            _nonStackData.push(series);
        }
    });

    return {
        _stackData,
        _nonStackData,
        stackIndices,
        nonStackseriesIndices: nonStackseriesIndices
    };
};
