// @zohocharts/zc/src/themes/colorpenciltheme.js
//$Id$
import { chartThemes } from '../core/main-zc-init';

export var colorpencil = function(userProp) {
    var colorpencilTheme = {
        tooltip: {
            borderTop: 20,
            backgroundColor: 'white',
            fontColor: 'black',
            opacity: 1
        },
        legend: {
            colorBox: {
                strokeWidth: 1
            }
            // "highlightEffect":{
            //     "selectedSeries":"flip"
            // },
            // "patterns" : ["hatchfill","crossStripesfill","squarefill","hatchLeftfill","wavefill","capfill","wavefill","starfill"]
            // "patterns" : ["bighatch","dot","crosscircle","crosscirclefill","hatchdashfill","checkerboard"],
            // "patternPallete" : "shapes-fill"
            // "patternPallete" : ['<rect x="0" y="0" width="10" height="10" style="fill: transparent;"></rect><path d="M 0 10 L 0 10 10 0 M 0 1 L 0 1 0 0 M 1 0 L 1 0 0 0 M 9 10 L 9 10 10 10 M 10 9 L 10 9 10 10" stroke="'+ "{{clr}}"+'" stroke-width="2" stroke-opacity="0.7"></path>']
        },
        chart: {
            plot: {
                renderer: {
                    mode: 'svg'
                },
                plotoptions: {
                    bar: {
                        border: {
                            show: true
                        }
                    },
                    area: {
                        strokeOpacity: 1,
                        strokeWidth: 1
                    },
                    bubblepie: {
                        strokeColor: null
                    },
                    pie: {
                        strokeColor: null,
                        datalabels: {
                            innerLabel: {
                                strokeColor: 'white',
                                strokeWidth: 6
                            }
                        }
                    },
                    funnel: {
                        strokeColor: null
                    },
                    pyramid: {
                        strokeColor: null
                    },
                    sunburst: {
                        strokeColor: null
                    },
                    dial: {
                        strokeColor: null
                    },
                    bullet: {
                        border: {
                            show: true,
                            size: null,
                            style: 'solid',
                            radius: 0,
                            color: null
                        }
                    },
                    gantt: {
                        border: {
                            show: true
                        }
                    },
                    scatter: {
                        marker: {
                            outerFillColor: null,
                            outerStrokeWidth: 1,
                            innerFillOpacity: 0,
                            innerStrokeOpacity: 0
                        },
                        hoveroptions: {
                            outerStrokeWidth: null
                        }
                    }
                }
            }
        }
    };
    return colorpencilTheme;
};
chartThemes.colorpencil = colorpencil;
