// @zohocharts/zc/src/themes/eleganttheme.js
//$Id$
//eleganttheme.js
import { chartThemes } from '../core/main-zc-init';
import { colorpencil } from './colorpenciltheme';
export var elegant = function(userProp) {
    // $ZCG.calcfontSize(denominator);
    var elegantTheme = {
        canvas: {
            title: {
                show: true,
                marginTop: 8,
                marginBottom: 8,
                fontColor: '#00bfdc',
                fontSize: 16,
                fontFamily: 'Arial'
            },
            subtitle: {
                show: false
            },
            border: {
                show: false
            },
            shadow: {
                show: false
            }
        },
        chart: {
            axes: {
                xaxis: {
                    label: {
                        fontColor: '#8f8f8f',
                        fontWeight: 'bold', //normal || bold || bolder || lighter || 100 || 200 || 300 || 400 || 500 || 600 || 700 || 800 || 900
                        fontFamily: null
                    },
                    ticklabel: {
                        fontColor: '#8f8f8f',
                        fontFamily: null,
                        fontWeight: 'bold' //normal || bold || bolder || lighter || 100 || 200 || 300 || 400 || 500 || 600 || 700 || 800 || 900
                    },
                    tickmark: {
                        color: '#c0c0c0',
                        size: 6
                    },
                    axisline: {
                        color: '#c0c0c0'
                    }
                },
                yaxis: [
                    {
                        label: {
                            fontColor: '#8f8f8f',
                            fontWeight: 'bold', //normal || bold || bolder || lighter || 100 || 200 || 300 || 400 || 500 || 600 || 700 || 800 || 900
                            fontFamily: null
                        },
                        ticklabel: {
                            fontColor: '#8f8f8f',
                            fontFamily: null,
                            fontWeight: 'bold', //normal || bold || bolder || lighter || 100 || 200 || 300 || 400 || 500 || 600 || 700 || 800 || 900
                            fontStyle: 'normal' //normal || italic || oblique
                        },
                        tickmark: {
                            size: 6
                        },
                        axisline: {
                            color: 'transparent'
                        }
                    }
                ]
            },
            plot: {
                plotoptions: {
                    bar: {
                        animation: {
                            type: 'vertical'
                        }
                    },
                    dial: {
                        strokeWidth: 0,
                        needle: {
                            strokeColor: '#505050'
                        }
                    }
                }
            }
        },
        legend: {
            colorPallete: {
                options: {
                    multicolor: 'BlueShades'
                }
            }
        },
        loader: {
            type: 'zBars',
            enabled: true,
            colors: ['#00c6e4', '#6ad4e4', '#a3dbe4', '#c3e0e4']
        }
    };
    return elegantTheme;
};

chartThemes.elegant = elegant;
