// @zohocharts/zc/src/themes/flatuitheme.js
//$Id$
//flatuitheme.js
import { chartThemes } from '../core/main-zc-init';
export var flatui = function(userProp) {
    // $ZCG.calcfontSize(denominator);
    var flatui = {
        canvas: {
            title: {
                show: true,
                marginBottom: 10,
                fontColor: '#22c48a',
                fontSize: 16,
                fontFamily: 'Roboto'
            },

            subtitle: {
                show: false,
                text: 'description',
                hAlign: 'center',
                marginLeft: 0,
                fontColor: '#CCCCCC'
            },
            border: {
                show: false,
                color: '#b0b7b9',
                radius: 0
            },
            shadow: {
                x: 0,
                y: 0
            }
        },

        chart: {
            axes: {
                xaxis: {
                    label: {
                        fontColor: '#979797',
                        fontFamily: null
                    },
                    ticklabel: {
                        fontColor: '#c0c0c0',
                        fontFamily: null
                    },
                    tickmark: {
                        size: 6
                    }
                },
                yaxis: [
                    {
                        label: {
                            fontColor: '#979797',
                            fontFamily: null
                        },
                        ticklabel: {
                            fontColor: '#979797',
                            fontFamily: null
                        },
                        tickmark: {
                            size: 6
                        },
                        axisline: {
                            color: 'transparent'
                        }
                    }
                ]
            },
            plot: {
                plotoptions: {
                    bar: {
                        levelMarker: {
                            color: 'rgba(0,0,0,0.2)'
                        }
                    },
                    pie: {
                        strokeColor: 'transparent'
                    }
                }
            }
        },
        legend: {
            fontColor: '#7b7d7e',
            colorPallete: {
                options: {
                    multicolor: 'flat-ui-colors'
                }
            }
        },
        loader: {
            enabled: true,
            type: 'zCircle2'
        }
    };
    return flatui;
};

chartThemes.flatui = flatui;
