// @zohocharts/zc/src/themes/penciltheme.js
//$Id$
//penciltheme.js
import { chartThemes } from '../core/main-zc-init';
export var pencil = function(userProp) {
    // $ZCG.calcfontSize(denominator);
    var pencilTheme = {
        canvas: {
            fontColor: '#cccccc',
            title: {
                marginBottom: 10,
                fontColor: '#666666',
                fontSize: 18,
                fontFamily: "'Gloria Hallelujah', cursive"
            },
            subtitle: {
                show: false
            },
            background: {
                color: 'transparent'
            },
            border: {
                show: false,
                style: 'dashed',
                radius: 0
            },
            shadow: {
                show: false
            }
        },
        chart: {
            effects: {
                choice: 'freeHand'
            },
            axes: {
                xaxis: {
                    label: {
                        fontColor: '#333333',
                        fontFamily: null
                    },
                    ticklabel: {
                        fontColor: '#666666',
                        fontFamily: null,
                        fontWeight: 'bold'
                    },
                    tickmark: {
                        color: '#c0c0c0',
                        size: 6
                    },
                    axisline: {
                        color: '#666666'
                    }
                },
                yaxis: [
                    {
                        label: {
                            fontColor: '#979797',
                            fontFamily: null
                        },
                        grid: {
                            color: '#d3d3d3'
                        },
                        ticklabel: {
                            fontColor: '#979797',
                            fontFamily: null,
                            fontWeight: 'bold'
                        },
                        tickmark: {
                            color: '#979797',
                            size: 6
                        },
                        axisline: {
                            color: 'transparent'
                        },
                        threshold: {
                            line: {
                                label: {
                                    text: ['Description comes here'],
                                    hAlign: 'right',
                                    fontFamily: null,
                                    fontSize: '12',
                                    fontColor: 'black'
                                }
                            },
                            range: {
                                fillColor: 'grey',
                                fillOpacity: 0.2,
                                strokeColor: 'black'
                            }
                        }
                    }
                ]
            },
            plot: {
                plotoptions: {
                    pie: {
                        strokeColor: 'rgba(0,0,0,0.4)'
                    }
                }
            }
        },
        legend: {
            fontColor: '#111111',
            colorPallete: {
                options: {
                    multicolor: 'graphite6'
                }
            },
            imagePallete: 'linesndots'
        },
        loader: {
            type: 'zBars',
            enabled: true,
            colors: ['#595959', '#767676', '#9d9d9d', '#bfbfbf']
        }
    };
    return pencilTheme;
};

chartThemes.pencil = pencil;
