// @zohocharts/zc/src/themes/transparenttheme.js
//$Id$
//transparenttheme.js
import { chartThemes } from '../core/main-zc-init';
export var transparent = function(userProp) {
    // $ZCG.calcfontSize(denominator);
    var theme = {
        canvas: {
            title: {
                show: false
            },
            subtitle: {
                show: false
            },
            background: {
                alpha: 0,
                color: 'white'
            },
            border: {
                show: false
            },
            shadow: {
                show: true,
                x: 0,
                y: 0
            }
        },
        chart: {
            axes: {
                xaxis: {
                    show: false
                },
                yaxis: [
                    {
                        show: false
                    }
                ]
            }
        },
        legend: {
            colorPallete: {
                options: {
                    multicolor: 'graphite6'
                }
            }
        }
    };
    return theme;
};

chartThemes.transparent = transparent;
