// @zohocharts/zc/src/themes/zotheme.js
//$Id$
//zotheme.js
import { chartThemes } from '../core/main-zc-init';
export var zo = function(userProp) {
    // $ZCG.calcfontSize(denominator);
    var zoTheme = {
        canvas: {
            title: {
                marginBottom: 10,
                fontColor: '#23a96e',
                fontSize: 16,
                fontFamily: 'Roboto',
                textShadow: '1px 1px 1px rgba(0,0,0,0.2)'
            },
            subtitle: {
                show: false,
                text: 'description',
                hAlign: 'center',
                fontColor: '#CCCCCC'
            },
            background: {
                color: '#2e2f31'
            },
            border: {
                show: false,
                size: 1,
                style: 'solid',
                radius: 0,
                color: '#b0b7b9'
            },
            shadow: {
                x: 0,
                y: 0
            }
        },
        chart: {
            axes: {
                xaxis: {
                    label: {
                        fontColor: '#979797',
                        fontFamily: null
                    },
                    ticklabel: {
                        fontColor: '#979797',
                        fontFamily: null
                    },
                    tickmark: {
                        color: '#3d3e40',
                        size: 6
                    },
                    axisline: {
                        color: '#3d3e40'
                    }
                },
                yaxis: [
                    {
                        label: {
                            fontColor: '#979797',
                            fontFamily: null
                        },
                        grid: {
                            color: '#3d3e40'
                        },
                        ticklabel: {
                            fontColor: '#979797',
                            fontFamily: null
                        },
                        tickmark: {
                            color: '#3d3e40',
                            size: 6
                        },
                        axisline: {
                            color: 'transparent'
                        },
                        threshold: {
                            line: {
                                label: {
                                    hAlign: 'right',
                                    fontFamily: null,
                                    fontSize: '12',
                                    fontColor: 'black'
                                }
                            },
                            range: {
                                fillColor: 'grey',
                                fillOpacity: 0.2,
                                strokeColor: 'black'
                            }
                        }
                    }
                ]
            },
            plot: {
                plotoptions: {
                    bar: {
                        levelMarker: {
                            color: 'rgba(0,0,0,0.2)'
                        },
                        datalabels: {
                            show: false
                        }
                    }
                }
            }
        },
        legend: {
            fontColor: '#7b7d7e',
            colorPallete: {
                options: {
                    multicolor: 'flat-ui-colors'
                }
            }
        },
        loader: {
            type: 'zBars',
            enabled: true
        }
    };
    return zoTheme;
};

chartThemes.zo = zo;
