// @zohocharts/zc/src/events/visualEffects/behaviour-mapping.js
import { highlighter_c } from './controllers/highlighter_c';
import { reversehighlight_c } from './controllers/reversehighlight_c';
import { tooltip_c } from './controllers/tooltip_c';
import { highlightTouch_c } from './controllers/highlight-touch_c';
import { hidetooltip_c } from './controllers/hidetooltip_c';
import { notes_c } from './controllers/notes_c';
import { reset_c } from './controllers/reset_c';
import { showCursor_C } from './controllers/cursor_c';
import { resethighlight_c, forceResetHighlight_c } from './controllers/resethighlight_c';
import { markerBehaviourMap } from '../../components/marker/marker-behaviour-mapping';

var behaviourMap = {};
behaviourMap._defaultPlotBehaviour = {
    mousemove: [
        {
            action: () => {
                return resethighlight_c;
            },
            allowifcustom: true,
            type: 'resetHighlight'
        },
        {
            action: () => {
                return highlighter_c;
            },
            type: 'highlight'
        },
        {
            action: () => {
                return showCursor_C;
            },
            type: 'cursor'
        },
        {
            action: () => {
                return reversehighlight_c;
            },
            type: 'reverseHighlight'
        },
        {
            action: () => {
                return tooltip_c;
            },
            allowDefault: true,
            type: 'tooltip'
        }
    ],
    click: [
        {
            action: () => {
                return hidetooltip_c;
            },
            type: 'hideTooltip'
        },
        {
            action: () => {
                return showCursor_C;
            },
            type: 'cursor',
            allowifcustom: conditionalDefault
        },
        {
            action: () => {
                return highlightTouch_c;
            },
            type: 'highlightOnTaphold'
        }
    ],
    doubleclick: [
        {
            action: () => {
                return hidetooltip_c;
            },
            type: 'hideTooltip'
        },
        {
            action: () => {
                return notes_c;
            },
            type: 'notes'
        },
        {
            action: () => {
                return showCursor_C;
            },
            type: 'cursor',
            allowDefault: conditionalDefault
        }
    ],
    mouseout: [
        {
            action: () => {
                return resethighlight_c;
            },
            type: 'resetHighlight',
            allowDefault: true
        },
        {
            action: () => {
                return reset_c;
            },
            type: 'resetTooltip',
            allowDefault: true
        }
    ],
    contextmenu: []
};
behaviourMap._defaultPlotBehaviour.touchmove = behaviourMap._defaultPlotBehaviour.mousemove;
behaviourMap._defaultPlotBehaviour.tap = behaviourMap._defaultPlotBehaviour.click;
behaviourMap._defaultPlotBehaviour.doubletap = behaviourMap._defaultPlotBehaviour.doubleclick;

export function setPlotDefaultBehaviour() {
    if (arguments.length === 2) {
        behaviourMap[arguments[0]] = arguments[1];
    } else {
        behaviourMap._defaultPlotBehaviour = arguments[0];
    }
}

export function getPlotDefaultBehaviour() {
    if (arguments.length) {
        let [name] = arguments;
        if (name) {
            return behaviourMap[name];
        }
    } else {
        return behaviourMap._defaultPlotBehaviour;
    }
}

function conditionalDefault(args) {
    if (this.chart.isAxisCategory || this.chart.isPolarAxisCategory) {
        return true;
    }
    return false;
}

setPlotDefaultBehaviour('marker', markerBehaviourMap);

export { conditionalDefault };
