// @zohocharts/zc/src/behaviours/scroll/brush-scroll-handler-mappings.js
import { ZCChartScrollListener } from '../listeners/zc-brush-scroll-listener';
import { BrushScrollDataFinder } from '../data-selection/chart/scroll-brush-data-finder';
import { SCROLL_BRUSH_ACTION } from '../visual-effects/chart/controllers/scroll-chart-by-brush';
import { Registry } from '../../core/zc-registry';
import { d3_set } from '../../core/d3-constants';
import { hideScrollTooltip } from '../visual-effects/chart/controllers/hide-tooltip';
import { ShowBrushResetButton } from '../visual-effects/chart/controllers/plot-reset-button';

let _chartScrollHandlers = {
    LISTENER: {
        start: ZCChartScrollListener,
        brush: ZCChartScrollListener,
        end: ZCChartScrollListener
    },
    DATA_FINDER: {
        start: BrushScrollDataFinder,
        brush: BrushScrollDataFinder,
        end: BrushScrollDataFinder
    },
    BEHAVIOUR: {
        // start: SCROLL_BRUSH_ACTION,
        brush: SCROLL_BRUSH_ACTION,
        end: [hideScrollTooltip, ShowBrushResetButton]
    },
    LISTENER_TYPES: d3_set(['start', 'brush', 'end']),
    SUPPORTED: chart => chart.isAxisCategory && !chart.isPolarAxisCategory
};

Registry.setComponent('_chartScrollHandlers', _chartScrollHandlers);

export { _chartScrollHandlers };
