// @zohocharts/zc/src/behaviours/scroll/chart-brush-scroll.js
//$Id$
import { Registry } from '../../core/zc-registry';
import { fixBrushDirection } from '../data-selection/chart/tools/selection-to-domain';
import { scale } from '../../core/main-zc-init';
import { mathAbs, mathFloor, mathMax, mathMin, UNDEFINED } from '../../core/zc-constants';
import { getCategoryStep, getOuterPaddingInPixel } from '../../axis/scale';
import { createElement } from '../../utils/zcsvgutils';
import { getMargin } from '../../axis/axisutils';
import { d3_ascending, d3_event, d3_scaleLinear } from '../../core/d3-constants';
import {
    clipOverview,
    getBrushExtent,
    getHandleMarkerSize,
    getScrollBoundRect,
    renderOverviewContext
} from '../visual-effects/chart/controllers/overview';
import { getBrushHandle } from '../visual-effects/brush-handles/handles';
import { defined, each, firstAndLastItem, getPropVal, invokeFunction, isArray, pick } from '../../utils/zcutils';
import { BrushComponentTooltip } from '../../legend/colorband/colorband-brush-tooltip';
import { getAutoTimeSubFunction, getFormattedValue, toFixedNumber } from '../../formatter/formatter';
import { MINIMUM_VISIBLE_POINTS } from '../behavioural-constants';
import { getMinimumVisibleArea } from '../plot-zoom/behaviour-constrainer';
import { getScrollSelection } from '../visual-effects/chart/controllers/scroll-domain';
import { domainToActualRange } from '../visual-effects/chart/tools/zoom-identity-by-domain';
import { applyPadding } from '../visual-effects/chart/controllers/scroll-chart-by-brush';
import { TIME_DATATYPE } from '../../core/datatype-constants';

let BrushD3 = Registry.getComponent('BrushD3'),
    BrushD3Options = Registry.getComponent('BrushD3Options');
/**
 *
 */
export class ChartBrushScroll {
    constructor(chart, type) {
        this.chart = chart;
        this.rotated = this.chart.axes.rotated;
        this.settings = chart.systemConf.chart.scroll;
        this.type = type;
    }

    initialize() {
        let scroll = this,
            chart = scroll.chart,
            scrollSettings = scroll.settings,
            orient = scroll.rotated ? 'left' : 'bottom',
            hasPlotZoom = getPropVal(chart.systemConf, 'chart.zoom.enabled') === true,
            applySmoothEffect = hasPlotZoom || scrollSettings.smooth,
            brushScale = (scroll.brushScale = scroll.getScale(chart, chart.plotarea, scroll.type, orient, !applySmoothEffect));
        let selectionArgs = {
            scrollType: this.type,
            scrollScale: brushScale,
            chartObj: chart
        };
        scrollSettings.height = invokeFunction(scrollSettings.height, scrollSettings, chart);
        let userSelection = isArray(scrollSettings.selectionRange)
                ? scrollSettings.selectionRange.slice().sort(d3_ascending)
                : invokeFunction(scrollSettings.selectionRange, selectionArgs) || null,
            resizerSettings = scrollSettings.resizer || {},
            sliderSettings = scrollSettings.slider || {},
            resizerEnabled = resizerSettings.enabled === true,
            initialDomain = brushScale.domain();

        if (!resizerEnabled && sliderSettings.enabled) {
            resizerEnabled = true;
            resizerSettings = sliderSettings;
            scroll.behaviour = 'slider';
        }

        scroll.updatePositionScale();

        scroll.minimumVisiblePoints = pick(scrollSettings.minimumVisiblePoints, MINIMUM_VISIBLE_POINTS);
        scroll.resizerSettings = resizerSettings;
        scroll.resizerEnabled = resizerEnabled;
        scroll.overviewEnabled = getPropVal(scrollSettings, 'overview.enabled') === true;
        scroll.isSmooth = applySmoothEffect;

        // since during zoom the padding is already there, no need to update the padding
        // #ZC2304 $minimumVisiblePoints
        if (!hasPlotZoom) {
            scroll.fixScaleOuterPadding();
        }

        let scrollSpace = chart.plotarea[scroll.rotated ? 'height' : 'width'],
            domain = scroll.getInitialDomain(),
            selection = userSelection || scroll.getScrollSelection(domain),
            scrollLength = mathAbs(selection[1] - selection[0]),
            isBrushEnabled = chart.systemConf.chart.brush.enabled,
            // scroll should be visible if zoom or resize is enabled
            isScrollNeed = scroll.resizerEnabled || hasPlotZoom || scrollSpace - scrollLength >= 1 || isBrushEnabled;

        if ((scroll.visible = isScrollNeed)) {
            scroll.initialSelection = selection;
            scroll.scrollDomain = domain;

            if (!scroll.isSmooth && !scroll.resizerEnabled) {
                scroll.visiblePointCount = domain.length;
            }

            // initial rendering optimization
            if (!scroll.isSmooth) {
                // #ZC2455 Final
                chart.viewPortExtent = resizerEnabled ? initialDomain : domain;
            }

            scroll.updatePlotAreaByScroll();
        }
    }

    fixScaleOuterPadding() {
        let self = this,
            brushScale = self.brushScale;

        if (!brushScale.bandwidth) {
            let domain = firstAndLastItem(self.getInitialDomain()),
                cType = self.chart.chartTypes[0].name,
                plotoptions = self.chart.systemConf.chart.plot.plotoptions[cType],
                outerPadding = defined(plotoptions.outerPadding) ? plotoptions.outerPadding : 0.02,
                diff = (domain[1] - domain[0]) * outerPadding,
                currentDomain = self.brushScale.domain();

            currentDomain = applyPadding(currentDomain, null, [diff, diff]);
            brushScale.domain(currentDomain);
            brushScale.originalDomain = currentDomain;
        }
    }

    updatePositionScale() {
        let self = this,
            chart = self.chart,
            range,
            _orient = chart.systemConf.chart.scroll.orient;

        if (_orient === 'right' || _orient === 'left') {
            range = [chart.plotarea.height, 0];
        } else {
            range = [0, chart.plotarea.width];
        }
        if (!_orient) {
            if (chart.axes.rotated) {
                range = [chart.plotarea.height, 0];
            } else {
                range = [0, chart.plotarea.width];
            }
        }

        self.scrollPositionScale = d3_scaleLinear()
            .range(range)
            .domain([0, 100]);
    }

    update() {
        if (this.chart.phase === 'redraw') {
            this.settings = this.chart.systemConf.chart.scroll;
            this.rotated = this.chart.axes.rotated;
            return;
        }

        if (this.visible) {
            let scroll = this,
                chart = scroll.chart,
                xScale = chart.axes.x.scale,
                originalDomain = chart.axes.x.baseScale.domain();

            scroll.updatePositionScale();

            chart.cache.scroll.selection = null;

            if (xScale.bandwidth) {
                chart.viewPortExtent = null;
            }
            scroll.scrollDomain = chart.cache.scroll.domain ? chart.cache.scroll.domain : scroll.scrollDomain; // scrollDomain is undefined -when legend filter made in racing chart as chart.cache.scroll.domain is undefined, and error produced on end of brush for undefined scrollDomain,thus set scrollDomain to scrollDomain itself if cache scroll domain is undefined

            // overview chart re-render
            scroll.render();

            // update the domain
            // chart.axes.x.baseScale.range(xScale.range());
            if (!chart.racing) {
                scroll.brushScale.originalDomain = originalDomain;
                xScale.originalDomain = originalDomain;
            }
        }
    }

    render() {
        let scroll = this,
            chart = scroll.chart,
            scrollSettings = scroll.settings,
            rotated = scroll.rotated,
            overviewSettings = scrollSettings.overview || {}; //TODO MOve to some other child class

        // #ZC2019
        scroll.parent = createElement('g', '.brush-scroll', scroll.visible ? [1] : [], chart.staticPaneSVG, 'g.legendarea');

        if (scroll.parent.empty()) {
            return;
        }
        let topBotOrient = this.settings.orient === 'bottom' || this.settings.orient === 'top';
        let leftRightOrient = this.settings.orient === 'right' || this.settings.orient === 'left';
        if (!this.settings.orient) {
            scroll.brushType = fixBrushDirection('x', rotated);
        } else {
            scroll.brushType = fixBrushDirection('x', (rotated && !topBotOrient) || (!rotated && leftRightOrient));
        }
        scroll.brushExtent = getBrushExtent(chart, scrollSettings);
        scroll.position = scroll.getPosition();

        let position = scroll.position;
        scroll.dimension = getScrollBoundRect(chart, scroll);
        scroll.parent.attr('transform', `translate(${position.x},${position.y})`);

        // set it initial phase
        scroll.initialized = false;
        // create the overview
        if (overviewSettings.enabled) {
            renderOverviewContext(chart, scroll.dimension, overviewChart => {
                clipOverview.call(scroll, overviewChart);
            });
        } else {
            scroll.addEvent(scroll.brushExtent);
        }
        scroll.isRedraw = true;
    }

    getScrollSelection(domain) {
        let scroll = this,
            chart = scroll.chart,
            behaviours = (chart.behaviouralEvents && chart.behaviouralEvents.eventManager.behaviours) || {},
            transform = scroll.transform || (behaviours.PLOT_ZOOM && behaviours.PLOT_ZOOM.instance.getTransform()),
            selectArea;

        if (scroll.lastScroll) {
            selectArea = scroll.lastScroll.map(scroll.scrollPositionScale);
        } else {
            selectArea = getScrollSelection(chart, domain, transform);
        }
        selectArea.sort(d3_ascending);

        // very first selection for normal scroll
        if (!scroll.isRedraw && !scroll.overviewEnabled && !scroll.resizerEnabled) {
            let reversed = chart.systemConf.chart.axes.xaxis.reversed,
                _r = scroll.brushScale.bandwidth ? scroll.brushScale.origRange[0] : scroll.brushScale(scroll.brushScale.originalDomain[0]);
            if (scroll.rotated) {
                if (reversed) {
                    selectArea[0] = _r;
                } else {
                    selectArea[1] = _r;
                }
            } else {
                if (reversed) {
                    selectArea[1] = _r;
                } else {
                    selectArea[0] = _r;
                }
            }
        }

        return selectArea;
    }

    addEvent(extent, domain) {
        let scroll = this,
            chart = scroll.chart,
            scrollSettings = scroll.settings,
            resizerSettings = scroll.resizerSettings,
            resizerEnabled = scroll.resizerEnabled,
            trackerStyles = scrollSettings.tracker,
            thumbStyles = scrollSettings.thumb,
            scrollCache = (chart.cache.scroll = chart.cache.scroll || {}),
            fixRadius = obj => obj.radius && (obj.rx = obj.ry = obj.radius),
            overviewOffset = chart.overview && [chart.overview.plotarea.left, chart.overview.plotarea.top],
            selectionArgs = {
                scrollType: this.type,
                scrollScale: scrollSettings.brushScale,
                chartObj: chart
            },
            userSelection = isArray(scrollSettings.selectionRange)
                ? scrollSettings.selectionRange.slice().sort(d3_ascending)
                : invokeFunction(scrollSettings.selectionRange, selectionArgs) || null,
            selectArea = userSelection || scroll.getScrollSelection(scrollCache.domain || domain || scroll.scrollDomain);

        // fix radius
        fixRadius(thumbStyles);
        fixRadius(trackerStyles);

        let options = new BrushD3Options({
            type: scroll.brushType,
            behaviour: scroll.behaviour || 'scroll',
            extent: extent,
            resizer: resizerEnabled && resizerSettings.type === 'rectangle',
            styles: {
                track: trackerStyles,
                selection: thumbStyles,
                slider: scrollSettings.slider
            },
            brushSelection: scrollSettings.selectionDomain,
            brushScale: scrollSettings.brushScale
        });

        let brush = (scroll.brushInstance = new BrushD3(options));

        //
        brush.bindTo(scroll.parent); //TODO add this to chartBrush

        // add the listeners after binding the brush
        each(scroll._hooks, (action, type) => scroll.brushInstance.addHook(type, action));

        if (resizerEnabled) {
            let userStyles = resizerSettings[resizerSettings.type] || {};
            fixRadius(userStyles);
            let handles = getBrushHandle(scroll.brushType, resizerSettings.type, extent, null, userStyles, overviewOffset);
            brush.addCustomHandle(handles);
            scroll.enableTooltip(resizerSettings);
        }

        // set it to true
        scroll.initialized = true;

        scroll.constrain(brush);

        // set the initial scroll selection
        d3_event(null);
        brush.setSelection(selectArea);

        return brush;
    }

    enableTooltip() {
        let self = this,
            resizerSettings = self.settings.resizer,
            tooltipOptions = resizerSettings.tooltip || {};

        if (tooltipOptions.enabled === true) {
            let chart = self.chart,
                columnInfo = chart.dataset.getColumn('x'),
                datatype = columnInfo.datatype,
                formatConf = columnInfo[datatype] || {},
                orient = self.rotated ? 'right' : 'bottom',
                dataFinder = i => {
                    let domain = [chart.viewPortExtent[0], chart.viewPortExtent[chart.viewPortExtent.length - 1]];
                    if (columnInfo.datatype === TIME_DATATYPE) {
                        let specifier = getPropVal(formatConf, 'tooltipformat.specifier') || getPropVal(formatConf, 'format.specifier');
                        if (specifier === 'auto') {
                            let subs = getAutoTimeSubFunction(mathAbs(domain[1] - domain[0])),
                                auto = { YEAR: 'DATE', MONTH: 'DATE' };
                            formatConf = {
                                subfunction: auto[subs.subFunction] || 'DATETIME'
                            };
                        }
                    }

                    if (defined(tooltipOptions.content)) {
                        return invokeFunction(chart, domain[i], columnInfo);
                    } else {
                        return getFormattedValue(formatConf, domain[i], datatype, chart, 'tooltip', UNDEFINED, '-');
                    }
                };

            self.tooltip = new BrushComponentTooltip(tooltipOptions, self.brushInstance, orient, dataFinder, self.chart);
        }
    }

    on(type, action) {
        // TODO remove this and find better approach
        this._hooks = this._hooks || {};
        this._hooks[type] = action;
    }

    getScale(chartObj, plotarea, type, orient, noOuterPadding) {
        return (chartObj.axes.x.scrollScale = scale.getScale(
            chartObj,
            chartObj.dataObject,
            plotarea,
            orient,
            0,
            chartObj.dataset,
            type,
            null,
            noOuterPadding,
            false
        ));
    }

    constrain(brush) {
        let self = this,
            chart = self.chart,
            minimumScrollSize = getMinimumVisibleArea(chart, 'x', self.minimumVisiblePoints),
            isOverviewEnabled = chart.scroll.overviewEnabled;

        let start;
        brush.addHook('start', e => {
            start = e.selection;
            side = lastMin = null;
        });

        brush.addHook('brush', e => {
            if ((e.mode === 'resize' || (isOverviewEnabled && e.mode === 'move')) && !self.isSmooth) {
                self.visiblePointCount = null;
            }
        });

        brush.addHook('end', e => {
            if ((e.mode === 'resize' ||  (e.mode === 'move')) && !self.isSmooth) {
                self.visiblePointCount = self.scrollDomain.length;
            }
        });

        // update the scroll selection
        let lastMin = null,
            side = null;
        brush.snap('brush', function(event) {
            let selection = event.selection;

            if (event.mode === 'move') {
                return selection;
            }

            if (!side) {
                side = +selection[0] !== +start[0] ? 'left' : 'right';
            }

            if (side === 'left') {
                selection[1] = mathMin(start[1], selection[1]);
            } else {
                selection[0] = mathMax(start[0], selection[0]);
            }

            let width = toFixedNumber(mathAbs(selection[1] - selection[0]), 2);

            if (side === 'left' && lastMin) {
                if (selection[0] >= start[1] - minimumScrollSize) {
                    return lastMin;
                }
            }

            if (side === 'right' && lastMin) {
                if (selection[1] <= start[0] + minimumScrollSize) {
                    return lastMin;
                }
            }

            if (width < minimumScrollSize) {
                if (side === 'right') {
                    return (lastMin = [selection[0], selection[0] + minimumScrollSize]);
                } else {
                    return (lastMin = [selection[1] - minimumScrollSize, selection[1]]);
                }
            }

            return selection;
        });
    }

    getInitialDomain() {
        return this.getCategoryPosScreen();
    }

    getInitialSelection(domain) {
        let self = this,
            chart = self.chart;
        return domainToActualRange(chart, domain, 0, 'x', false, true, self.brushScale);
    }

    getCategoryPosScreen() {
        let self = this,
            chartObj = self.chart,
            scrollData = this.settings,
            plotarea = chartObj.plotarea;
        let defaultItemSize = 15;
        var numVisibleCategories = scrollData.visibleCategories;
        var cType = chartObj.chartTypes[0].name;
        var padding =
            cType == 'gantt' || cType == 'bullet' || cType == 'bar'
                ? chartObj.systemConf.chart.plot.plotoptions[cType].padding || 0.2
                : 0.2;
        var outerPadding = chartObj.systemConf.chart.plot.plotoptions[cType].outerPadding || 0.02;
        var categoryPosValue;

        if (self.resizerEnabled) {
            let xCat = chartObj.dataObject.getCategories('x', 0, false);
            xCat = xCat && xCat.length ? xCat : chartObj.dataObject.findXCategories();

            // single domain
            self.isSingleDomain = xCat.length === 1;

            return self.brushScale.domain();
        } else if (numVisibleCategories) {
            numVisibleCategories = mathMax(self.minimumVisiblePoints, scrollData.visibleCategories);
            categoryPosValue = this.getCategoryPosByCount(plotarea, padding, outerPadding, numVisibleCategories);
        } else {
            categoryPosValue = this.getCategoryPosByThickness(
                plotarea,
                padding,
                outerPadding,
                scrollData.categoryThickness || defaultItemSize
            );
        }
        return categoryPosValue;
    }

    getActualScreenValue(scrollScale, index, value) {
        if (scrollScale.bandwidth) {
            return scrollScale.ordinalRange[index];
        } else {
            return scrollScale(value);
        }
    }

    getScrollBarLength(scrollScale, categoryPos, userSelection) {
        let axisRotated = this.rotated;
        if (isArray(userSelection)) {
            return userSelection[1] - userSelection[0];
        }
        if (scrollScale.bandwidth) {
            if (axisRotated) {
                return scrollScale.range()[0] - categoryPos;
            } else {
                return categoryPos + scrollScale.bandwidth();
            }
        } else {
            if (axisRotated) {
                return scrollScale.range()[0] - categoryPos;
            } else {
                return categoryPos;
            }
        }
    }

    getCategoryPosByCount(plotarea, padding, outerPadding, numVisibleCategories) {
        var self = this,
            chart = self.chart,
            scrollScale = self.chart.axes.x.scrollScale,
            xCat = chart.dataObject.getCategories('x', 0, false),
            xCat = xCat && xCat.length ? xCat : chart.dataObject.findXCategories(),
            actNumCategories = xCat.length,
            filterByIndex = chart.dataObject.isWaterfallEnabled,
            visibleCategoriesCount = mathMin(numVisibleCategories, actNumCategories),
            behaviours = chart.behaviouralEvents && chart.behaviouralEvents.eventManager.behaviours,
            brushOptions = chart.systemConf.chart.brush;

        // make a indicator
        self.allCategoriesVisible = visibleCategoriesCount >= actNumCategories;

        // water fall domain
        if (filterByIndex) {
            xCat = scrollScale.domain();
        }

        // when all ticks are visible make return the actual domain
        if (self.allCategoriesVisible) {
            self.lastScroll = null;
            if (chart.cache.scroll) {
                chart.cache.scroll.domain = null;
            }

            if (scrollScale.bandwidth && !self.isSmooth && chart.phase === 'redraw') {
                chart.viewPortExtent = xCat;
            }

            return self.brushScale.domain();
        }

        // redraw same domain maintain
        if (self.chart.cache.scroll && self.chart.cache.scroll.domain) {
            // if the number of $visibleCategories changes maintain that
            let lastDomain = self.chart.cache.scroll.domain;
            if (scrollScale.bandwidth) {
                // #ZC2318 when brush is enabled use the last domain
                if (behaviours.PLOT_BRUSH && isArray(brushOptions.x)) {
                    return lastDomain;
                }
                if (lastDomain.length !== visibleCategoriesCount) {
                    // reset the last scroll and take from new domain
                    self.lastScroll = null;
                    chart.cache.scroll.domain = null;
                    return getItemsFromArrayByCount(xCat, xCat.indexOf(lastDomain[0]), visibleCategoriesCount);
                }
            }
            return self.chart.cache.scroll.domain;
        }

        let visibleCategories = xCat.slice(0, visibleCategoriesCount);

        if (scrollScale.bandwidth) {
            return visibleCategories;
        } else {
            return firstAndLastItem(visibleCategories);
        }
    }

    getCategoryPosByThickness(plotarea, padding, outerPadding, thickness) {
        var scrollScale = this.chart.axes.x.scrollScale,
            baseValue = this.rotated ? plotarea.height : plotarea.width;
        var step = getCategoryStep(thickness, padding);
        var numVisibleCategories = this.getVisibleNumCategories(this.chart, plotarea, outerPadding, step);

        if (scrollScale && scrollScale.bandwidth) {
            outerPadding = getOuterPaddingInPixel(scrollScale.range(), numVisibleCategories, padding, outerPadding, step);
        } else {
            outerPadding = 2 * (step * outerPadding);
        }
        var numVisibleCategories = this.getVisibleNumCategories(this.chart, plotarea, outerPadding, step);

        var categoryPosValue = this.getCategoryPosByCount(plotarea, padding, outerPadding, numVisibleCategories);
        return categoryPosValue;
    }

    getVisibleNumCategories(chartObj, plotarea, outerPadding, step) {
        return this.rotated ? mathFloor((plotarea.height - outerPadding) / step) : mathFloor((plotarea.width - outerPadding) / step);
    }

    updatePlotAreaByScroll(plotArea = this.chart.plotarea) {
        let chartScroll = this,
            scrollData = chartScroll.settings,
            axisRotated = chartScroll.rotated,
            markerSize = getHandleMarkerSize(chartScroll, getBrushExtent(chartScroll.chart, scrollData)),
            scrollSpan = scrollData.height + markerSize;

        let { left, top } = plotArea,
            margin = getMargin(scrollData, 0),
            scrollOrient = scrollData.orient;
        if (defined(scrollData.x) && defined(scrollData.y)) {
            left = invokeFunction(scrollData.x, chartScroll, this.chart);
            top = invokeFunction(scrollData.y, chartScroll, this.chart);
        } else if (scrollOrient) {
            if (scrollOrient === 'left') {
                plotArea.width -= scrollSpan + margin.left + margin.right;
                plotArea.left += scrollSpan + margin.left + margin.right;
                left += margin.left;
            } else if (scrollOrient === 'bottom') {
                plotArea.height -= scrollSpan + margin.top + margin.bottom;
                plotArea.bottom += scrollSpan + margin.top + margin.bottom;
                top += plotArea.height;
                top += margin.top;
            } else if (scrollOrient === 'right') {
                plotArea.width -= scrollSpan + margin.left + margin.right;
                left += plotArea.width;
                left += margin.left;
            } else if (scrollOrient === 'top') {
                plotArea.height -= scrollSpan + margin.top + margin.bottom;
                plotArea.bottom += scrollSpan + margin.top + margin.bottom;
                plotArea.top += scrollSpan + margin.top + margin.bottom;
                top += margin.top;
            }
        } else {
            if (axisRotated) {
                plotArea.width -= scrollSpan + margin.left + margin.right;
                left += plotArea.width;
                left += margin.left;
            } else {
                plotArea.height -= scrollSpan + margin.top + margin.bottom;
                plotArea.bottom += scrollSpan + margin.top + margin.bottom;
                top += plotArea.height;
                top += margin.top;
            }
        }
        //when strokeColor is there for thumb if height is 40, thumb's height exceeds 40 and gets cropped at bottom(if border is enabled false and marginbottom is not there)
        //inorder to prevent thumb from getting cropped due to strokeColor - 1px is subtracted from the y position of scroll
        top = scrollData.thumb.strokeColor ? top - 1 : top;
        chartScroll._position = { x: left, y: top };
    }

    getPosition() {
        let plotArea = this.chart.plotarea;
        let topBotOrient = this.settings.orient === 'bottom' || this.settings.orient === 'top';
        let leftRightOrient = this.settings.orient === 'right' || this.settings.orient === 'left';

        if (defined(this.settings.x) && defined(this.settings.y)) {
            return {
                x: this._position.x,
                y: this._position.y
            };
        } else if (this.settings.orient) {
            if ((this.chart.axes.rotated && !topBotOrient) || (!this.chart.axes.rotated && leftRightOrient)) {
                return {
                    x: this._position.x,
                    y: plotArea.top
                };
            } else {
                return {
                    x: plotArea.left,
                    y: this._position.y
                };
            }
        } else {
            if (this.chart.axes.rotated) {
                return {
                    x: this._position.x,
                    y: plotArea.top
                };
            } else {
                return {
                    x: plotArea.left,
                    y: this._position.y
                };
            }
        }
    }

    destroy() {
        this.chart.overview && this.chart.overview.destroy();
        this.brushInstance && this.brushInstance.destroy();
        this.parent.remove();
    }
}

/**
 *
 * @param array
 * @param fromIndex
 * @param itemCount
 * @return {*}
 */
function getItemsFromArrayByCount(array, fromIndex, itemCount) {
    let rightAvailable = array.length - fromIndex;
    if (itemCount > rightAvailable) {
        let fromLeft = Math.max(fromIndex - (itemCount - rightAvailable), 0);
        return array.slice(fromLeft);
    } else {
        return array.slice(fromIndex, fromIndex + itemCount);
    }
}

Registry.setComponent('ChartBrushScroll', ChartBrushScroll);
