// @zohocharts/zc/src/behaviours/scroll/scrollOptions.js
import { defaultTheme as options } from '../../core/defaulttheme';
import { POINTER } from '../../core/zc-constants';

options.scroll = function() {
    return {
        x: null,
        y: null,
        orient: null,
        enabled: true,
        categoryThickness: null,
        visibleCategories: null,
        height: 10,
        // marginTop: 5,
        // marginBottom: 5,
        // marginLeft: 5,
        // marginRight: 5,
        thumb: {
            fillColor: '#b3b6b7',
            fillOpacity: 0.3,
            strokeColor: 'black',
            strokeOpacity: 0.3,
            cursor: POINTER
        },
        resizer: {
            enabled: false,
            type: 'lever',
            tooltip: {
                backgroundColor: '#b3b6b7',
                borderColor: 'rgb(51, 51, 51)',
                fontSize: 14
            }
        },
        tracker: {
            fillColor: '#ECF0F1',
            cursor: POINTER
        },
        overview:{
            show: true,
            renderMode: "canvas"
        }
    };
};

export var _scrollOptions = options.scroll;
