// @zohocharts/zc/src/waterfall/comparisonLine.js
//$Id$
//comparisonLine.js
import { svg_plot_renderer } from '../core/main-zc-init';
import { getAnimationConf } from '../utils/animationutils';
import { mergeJSON, bbox, isPositive } from '../utils/zcutils';
import { appendEle, applyFont } from '../utils/zcsvgutils';
import { d3_line } from '../core/d3-constants';
import { utils, getValByPriority } from '../utils/zcutils';
import { getFormattedValue } from '../formatter/formatter';
import { mathMax, mathMin, mathAbs } from '../core/zc-constants';
import { getDashArray } from '../utils/dashstyles';
import { NUMERIC_DATATYPE } from '../core/datatype-constants';
import { addMarkerPointer } from '../bar/markerPointer';
import { DataProcessor } from '../core/data-processor';
import { $ZCG } from '../core/zcglobal';

export * from './waterfall-utils';

svg_plot_renderer.waterfallHelpers = function() {};
svg_plot_renderer.waterfallHelpers.drawComparisonLine = function(group, data, config) {
    var chartObj = config.chart,
        points = data,
        comparatorData = getDataForComparison(points, config),
        plotoptions = config.plotoptions,
        commonPlotOptions = config.commonPlotOptions,
        animationOptions = getAnimationConf(commonPlotOptions, plotoptions, chartObj, 'waterfall'),
        isAnimationEnable = animationOptions.enabled,
        comparisonConfig =
            plotoptions && plotoptions.comparisonLine
                ? mergeJSON(commonPlotOptions.comparisonLine, plotoptions.comparisonLine, true)
                : commonPlotOptions.comparisonLine;
    config.animationOptions = animationOptions;

    var d3Line = d3_line()
        .x(function(d, i) {
            return d[0];
        })
        .y(function(d, i) {
            return d[1];
        });
    var lineG = appendEle(group, 'g', comparisonConfig.enabled ? [1] : [], '.', 'class', 'zccomparisonLines');
    var line = appendEle(lineG, 'path', comparatorData, '.', 'class', 'lines', null, function(d) {
        return d.y;
    });

    line.attr('d', function(d, i) {
        return d3Line(d);
    })
        .styles({
            fill: 'none',
            stroke: comparisonConfig.strokeColor || 'black',
            'stroke-width': comparisonConfig.strokeWidth || 1,
            'stroke-dasharray': getDashArray(comparisonConfig.dashStyle, comparisonConfig.strokeWidth || 1),
            'stroke-opacity': 0
        })
        .transition()
        .duration(isAnimationEnable ? animationOptions.duration : 0)
        .styles({
            'stroke-opacity': comparisonConfig.strokeOpacity || 1
        });
    addMarkerPointer(group, data, comparatorData, comparisonConfig, config);
    addText(group, comparatorData, comparisonConfig, config);
};

function addText(group, points, comparisonConfig, config) {
    var chartObj = config.chart,
        systemConf = chartObj.systemConf,
        axisrotated = config.axisrotated,
        isAnimationEnable = config.animationOptions.enabled,
        padding = 10,
        labelConfig = comparisonConfig.label || {},
        labelPos = labelConfig.labelPos,
        labelSize = utils.getVal(getValByPriority([labelConfig, systemConf.chart, systemConf.canvas, $ZCG], 'fontSize')),
        labelFixedPos = labelPos == 'fixedTop' ? (axisrotated ? chartObj.plotarea.width : labelSize) : null;
    var defaultFormatConf = {
        format: {
            decimalPlaces: '2',
            signEnabled: true,
            suffix: '%'
        }
    };
    var labelG = appendEle(group, 'g', comparisonConfig.enabled ? [1] : [], '.', 'class', 'zccomparisonLabels');
    var label = appendEle(labelG, 'text', points, '.', 'class', 'labels', null, function(d) {
        return d.y;
    });
    label
        .attrs({
            x: function(d, i) {
                return axisrotated ? labelFixedPos || d[1][0] + padding : d[0][0] + (d[2][0] - d[0][0]) / 2;
            },
            y: function(d, i) {
                return axisrotated ? d[1][1] + (d[2][1] - d[1][1]) / 2 : labelFixedPos || d[1][1] - 10;
            }
        })
        .text(function(d, i) {
            return getFormattedValue(defaultFormatConf, d.y, NUMERIC_DATATYPE, chartObj);
        })
        .styles({
            'text-anchor': axisrotated ? (labelPos == 'fixedTop' ? 'end' : 'start') : 'middle',
            opacity: 0
        })
        .transition()
        .duration(isAnimationEnable ? config.animationOptions.duration : 0)
        .styles({
            opacity: 1
        });
    applyFont(labelG, [labelConfig, systemConf.chart, systemConf.canvas, $ZCG]);
}
function getDataForComparison(points, config) {
    var chartObj = config.chart,
        renderer = config.renderer,
        dataset = chartObj.dataset,
        seriesindex = config.seriesindex,
        yaxiscolumnorder = config.yaxiscolumnorder,
        stacked = config.stacked,
        processedStackedData = chartObj.dataObject.waterfall.layout._keys,
        isCatNullY = DataProcessor.helpers.isCatNull(dataset, 'y', yaxiscolumnorder),
        yAccessor = getY,
        count = points.length - 1,
        datalabelPadding,
        barSize = config.axisrotated
            ? renderer.getParamValue('height', points[0], seriesindex, null, renderer.commonOption)
            : renderer.getParamValue('width', points[0], seriesindex, null, renderer.commonOption),
        mSize = mathMin(barSize / 10, 1) * 10,
        stackExtent = config.chart.dataObject.stackedChart ? config.chart.dataObject.waterfall.stackExtent[0] : null,
        cascadedPoints = getCascadedData(points),
        cascadeCounter = 0,
        comparatorPathData = [];

    points.map(function(d, i) {
        var point_now = points[i];

        if (point_now.isCascaded && points[i + 1]) {
            var cascaded_next = getNextCascadedPoint(points, i + 1, config);
            if (cascaded_next) {
                var intermediateMaxPt = cascaded_next.maxPt,
                    f = getTransValue(point_now, 'x', config, stackExtent),
                    intermediate_x = getTransValue(intermediateMaxPt, 'x', config, stackExtent),
                    intermediate_y = getTransValue(intermediateMaxPt, 'y', config, stackExtent),
                    point_now_x = getTransValue(point_now, 'x', config, stackExtent),
                    point_now_y = getTransValue(point_now, 'y', config, stackExtent),
                    point_now_width = getTransValue(point_now, 'width', config, stackExtent),
                    point_now_height = getTransValue(point_now, 'height', config, stackExtent),
                    point_now_unshared = point_now.index == 0 || point_now.index == points.length - 1,
                    point_next = cascaded_next.point,
                    point_next_x = getTransValue(point_next, 'x', config, stackExtent),
                    point_next_y = getTransValue(point_next, 'y', config, stackExtent),
                    point_next_width = getTransValue(point_next, 'width', config, stackExtent),
                    point_next_height = getTransValue(point_next, 'height', config, stackExtent),
                    point_next_unshared = point_next.index == 0 || point_next.index == points.length - 1,
                    movingSteps = getStepCoordinates(
                        {
                            x: point_now_x,
                            y: point_now_y,
                            width: point_now_width,
                            height: point_now_height,
                            unshared: point_now_unshared
                        },
                        {
                            x: point_next_x,
                            y: point_next_y,
                            width: point_next_width,
                            height: point_next_height,
                            unshared: point_next_unshared
                        },
                        {
                            x: intermediate_x,
                            y: intermediate_y
                        },
                        mSize,
                        config
                    );
                if (stacked) {
                    var point_now_stkd = processedStackedData[point_now.index],
                        point_next_stkd = processedStackedData[point_next.index];
                    movingSteps.y =
                        ((point_next_stkd.stackedValue - point_now_stkd.stackedValue) /
                            mathMax(point_now_stkd.stackedValue, point_next_stkd.stackedValue)) *
                        100;
                } else {
                    movingSteps.y = ((point_next.y - point_now.y) / point_now.y) * 100;
                }

                point_next.percent = movingSteps.y;
                if (cascadeCounter == 0 && i != 0) {
                    var initY = renderer.getParamValue('y', points[0], seriesindex, null, renderer.commonOption);
                    point_now.percent = ((point_now.y - points[0].y) / point_now.y) * 100;
                }
                comparatorPathData.push(movingSteps);
            }
            cascadeCounter++;
        }
    });
    return comparatorPathData;
}

function getStepCoordinates(source, target, intermediate, mSize, config) {
    var axisrotated = config.axisrotated,
        stacked = config.chart.dataObject.stacked,
        datalabelPadding = config.chart.dataObject.waterfall.datalabelPadding,
        tinyAdjustmentPadding = 30,
        x_target = mathMax(intermediate.x, stacked ? target.x : target.x + target.width), //TODO comment
        y_target = mathMin(intermediate.y, target.y),
        width_source = source.unshared ? source.width / 2 : source.width / 2 + source.width / 4,
        width_target = target.unshared ? target.width / 2 : target.width / 4,
        height_source = source.unshared ? source.height / 2 : source.height / 4,
        height_target = target.unshared ? target.height / 2 : target.height / 2 + target.height / 4,
        step_1,
        step_2,
        step_3,
        step_4;

    if (axisrotated) {
        (step_1 = [source.x + source.width + datalabelPadding, source.y + height_source]),
            (step_2 = [x_target + datalabelPadding + tinyAdjustmentPadding, source.y + height_source]),
            (step_3 = [x_target + datalabelPadding + tinyAdjustmentPadding, target.y + height_target]),
            (step_4 = [target.x + target.width + datalabelPadding, target.y + height_target]);
    } else {
        (step_1 = [source.x + width_source, source.y - datalabelPadding]),
            (step_2 = [source.x + width_source, y_target - datalabelPadding - tinyAdjustmentPadding]),
            (step_3 = [target.x + width_target, y_target - datalabelPadding - tinyAdjustmentPadding]),
            (step_4 = [target.x + width_target, target.y - datalabelPadding]);
    }

    var moves = [step_1, step_2, step_3, step_4];
    return moves;
}

function getNextCascadedPoint(points, index, config) {
    var point_now = points[index],
        stackExtent = config.chart.dataObject.stackedChart ? config.chart.dataObject.waterfall.stackExtent[0] : null,
        maxY = 0,
        maxPoint = point_now;
    if (!point_now) {
        return;
    }
    while (!point_now.isCascaded) {
        var point_now_y = getY(point_now, stackExtent);
        if (point_now_y > maxY) {
            maxPoint = point_now;
            maxY = point_now_y;
        }
        point_now = points[++index];
        if (!point_now) {
            return;
        }
    }
    return { point: point_now, maxPt: maxPoint };
}
function getCascadedData(points) {
    var cascadedPoints = points.filter(function(d, i) {
        return d.isCascaded;
    });
    return cascadedPoints;
}

function getY(point, stackExtent) {
    if (stackExtent) {
        return getStackExtentValue(point, stackExtent);
    }
    return point.y + point.y0;
}

function getStackExtentValue(point, stackExtentList) {
    /*
     extent[0] = category index (for same category there may be two set of values, one for positive, one for negative)
     extent[1] = Display value
     extent[2] = cumulative screen value
    */

    var itemPos = point.index,
        stackExtent_now = stackExtentList.filter(function(d, i) {
            return d[0] == itemPos;
        }), // returns positive and negative list for the current category
        curEntry = stackExtent_now[0]; //Take only the first position of the list;
    // Always 0th position contains positive Details, 1st position contains negative info
    if (!curEntry || !curEntry.length) {
        return 0;
    }
    var actualDisplayVal = curEntry[1],
        actualExtentVal = curEntry[2],
        overallCumulativeVal = isPositive(actualDisplayVal) ? actualExtentVal : actualExtentVal + mathAbs(actualDisplayVal);

    return overallCumulativeVal;
}
//TODO COMMENT all params clearly
function getTransValue(point, type, config, stackExtent) {
    var renderer = config.renderer,
        axisrotated = config.axisrotated,
        seriesindex = config.seriesindex;

    if (stackExtent) {
        if (axisrotated && type == 'width') {
            return 0;
        }
        var actualExtent = getStackExtentValue(point, stackExtent);
        return (axisrotated && type == 'x') || (!axisrotated && type == 'y')
            ? renderer.commonOption.yscale(actualExtent)
            : renderer.getParamValue(type, point, seriesindex, null, renderer.commonOption);
    } else {
        return renderer.getParamValue(type, point, seriesindex, null, renderer.commonOption);
    }
}

export function getPreCalculatedLabelSize(chartObj, data) {
    var commonPlotOptions = chartObj.systemConf.chart.plot.plotoptions.waterfall,
        axisrotated = chartObj.axes.rotated,
        systemconf = chartObj.systemConf,
        labelSize = 0,
        padding = 10,
        element = appendEle(chartObj.staticPaneSVG.selectAll('g.labelholder'), 'text', [1], '.', 'class', 'dummyText');
    chartObj.seriesdata.forEach(function(d, i) {
        if (d.plotoptions || !labelSize) {
            var plotoptions = d.plotoptions,
                comparisonConfig =
                    plotoptions && plotoptions.comparisonLine
                        ? mergeJSON(commonPlotOptions.comparisonLine, plotoptions.comparisonLine, true)
                        : commonPlotOptions.comparisonLine,
                labelConfig = comparisonConfig.label;
            element.text('800%').style('fill', 'none');
            applyFont(element, [labelConfig, systemconf.chart, systemconf.canvas, $ZCG]);
            var clientRect = bbox(element.node());
            labelSize = mathMax(axisrotated ? clientRect.width : clientRect.height, labelSize);
        }
    });
    element.remove();
    return labelSize + padding;
}

export function checkComparisonLineEnabled(chartObj, index) {
    var data = chartObj.dataObject.getEncode('y', index).seriesData,
        commonPlotOptions = chartObj.systemConf.chart.plot.plotoptions.waterfall;

    var enabled = data.some(function(series) {
        var plotoptions = series.plotoptions,
            comparisonConfig =
                plotoptions && plotoptions.comparisonLine
                    ? mergeJSON(commonPlotOptions.comparisonLine, plotoptions.comparisonLine, true)
                    : commonPlotOptions.comparisonLine;

        return comparisonConfig.enabled;
    });
    return enabled;
}

export var _waterfallComparisonLine = svg_plot_renderer.waterfallHelpers.drawComparisonLine;
