// @zohocharts/zc/src/waterfall/waterfallOptions.js
import { defaultTheme as options } from '../core/defaulttheme';
options.waterfall = function() {
    return {
        outerPadding: 0.04,
        fillOpacity: 0.9,
        maxBandWidth: 200,
        multiColoring: false,
        border: {
            show: false,
            size: null,
            style: 'solid',
            radius: 0,
            color: null
        },
        stacked: {
            enabled: false,
            showAs: 'values' //values,percent
        },
        connector: {},
        comparisonLine: {},
        cascadeCategories: {
            dataindex: 2
        },
        risingCategories: {},
        fallingCategories: {},
        datalabels: {
            showAs: 'y', //x,y,percent
            stackLabels: {
                show: false,
                showAs: 'y' //y,percent
            }
        },
        animation: {
            type: 'verticalAll' //verticalAll || vertical
        },
        gradients: {
            //linear || none
            options: {
                linear: {
                    y2: 50
                }
            }
        }
    };
};

export var _waterfallOptions = options.waterfall;
