// @zohocharts/zc/src/legend/colorband/colorband-brush-tooltip.js
//$Id$
//brush-component-tooltip.js
import { SVGTooltip } from '../../tooltip/svg-tooltip';
import { attachevents } from '../../events/eventActions';
import { d3_event, d3_select } from '../../core/d3-constants';
import { invokeFunction, isObjectPropDefined } from '../../utils/zcutils';
import { getSVGTooltipPosition } from '../events/visual-effects/highlight-tooltip';
import { translateString } from '../../utils/svg-transforms';

var defaultOptions = {
    padding: 5,
    margin: 5
};

/**
 *
 * @param tooltipOptions
 * @param brushInstance
 * @param orient
 * @param dataFinder
 * @param customActions
 * @param chart
 * @constructor
 */
export function BrushComponentTooltip(tooltipOptions, brushInstance, orient, dataFinder, chart, customActions, globalFont, options = {}) {
    let handles = brushInstance.customHandles,
        { systemConf } = chart || {},
        _options = {
            config: { ...defaultOptions, ...tooltipOptions, orient },
            container: handles,
            globalFont: globalFont || [systemConf.chart, systemConf.canvas]
        },
        spec = isObjectPropDefined(brushInstance.options, 'component.spec', true);

    this.dataFinder = dataFinder;
    this.chart = chart;
    var tooltipObj = (this.super = new SVGTooltip(_options));
    this.container = tooltipObj.container;
    this.boundBoxEle = tooltipObj.boundBoxEle;
    // this.textEle = tooltipObj.textEle;
    this.tooltipOptions = tooltipObj.config;
    this.customActions = customActions || {};
    this.options = options;
    this.positionTooltip();
    if (!spec || !(spec.encode === 'clr' && isObjectPropDefined(spec.config, 'events.enabled', true) === false)) {
        this.addHoverEvents(brushInstance.parent_group);
    }
}

BrushComponentTooltip.prototype.show = function() {
    var self = this,
        count = 0;
    self.super.show(() => self.dataFinder(count++));
    if (self.customActions.show) {
        self.customActions.show();
    }
};

BrushComponentTooltip.prototype.hide = function(forced) {
    var self = this;
    self.super.hide(forced);
    if (self.customActions.hide) {
        self.customActions.hide();
    }
};

BrushComponentTooltip.prototype.positionTooltip = function() {
    var tooltipOptions = this.tooltipOptions,
        { orient } = this.super.config,
        { translate, boxElement = ele => ele, isHorizontal } = this.options;

    this.container.attr('transform', function() {
        let handleElement = boxElement(d3_select(this.parentNode)),
            position = getSVGTooltipPosition(
                handleElement,
                tooltipOptions,
                invokeFunction(isHorizontal, orient),
                invokeFunction(translate, handleElement)
            );

        return translateString(position);
    });
};

BrushComponentTooltip.prototype.addHoverEvents = function(element) {
    var self = this,
        onSelection = true,
        mouseHandlers = {
            mousemove: function(e) {
                var className = d3_select(d3_event(e).srcElement || d3_event().target).attr('class');
                if (className !== 'overlay') {
                    self.show();
                    onSelection = true;
                } else if (onSelection) {
                    self.hide();
                    onSelection = false;
                }
            },
            mouseleave: function() {
                self.hide();
                onSelection = false;
            }
        };
    attachevents(element, mouseHandlers);
};
