// @zohocharts/zc/src/legend/colorband/colorband-brush-action.js
import { isFunction } from '../../utils/function-utils';
import { NULL } from '../../core/zc-constants';
import { isArray, jsonStringify } from '../../utils/zcutils';
import { d3_ascending } from '../../core/d3-constants';
import { getRange, snapHandlePosition, styleHandles } from './colorband-brush-helpers';
import { updateTrackerClip } from './colorband-track';

/**
 *
 * @param chartObj
 * @param colorBandBrush
 * @return {function(...[*]=)}
 */
function filterRange(chartObj, colorBandBrush) {
    let previousRange = [],
        options = colorBandBrush.helperOptions,
        config = options.config,
        brushInstance = colorBandBrush.brushInstance;

    return function() {
        // update the track
        updateTrackerClip(options, brushInstance.parent_group, brushInstance, chartObj);

        let eventInitialized = isFunction(chartObj.legend.component.highlightItem),
            selection = brushInstance.getSelection(),
            range = getRange(options, selection);

        // style the handle
        styleHandles(options, colorBandBrush.slider, range);

        if (!eventInitialized || options.singleDomainContinuous) {
            return;
        }

        config.filter.range = range;
        chartObj.legend.component.highlightItem(NULL, false, true);

        if (jsonStringify(range) !== jsonStringify(previousRange)) {
            // #ZC1369 addRemoveSeries only if previousExtent and extent are diff
            chartObj.addRemoveSeries();
        }

        // update the tooltip
        colorBandBrush.tooltipComponent.show();

        previousRange = range;
    };
}

/**
 *
 * @param chartObj
 * @param colorBandBrush
 * @return {function(...[*]=)}
 */
function snapRange(chartObj, colorBandBrush) {
    let options = colorBandBrush.helperOptions,
        brushInstance = colorBandBrush.brushInstance,
        previousSelection = [];

    return function(event, brushed) {
        chartObj.cache.search = {}; //reset cache on end of the legend color band brush
        if (colorBandBrush.tooltipComponent) {
            colorBandBrush.tooltipComponent.hide();
        }

        // #ZC2276 $selection && $previousSelection
        let selection = brushInstance.getSelection(),
            snapSelection = snapHandlePosition(options, selection),
            isSameSelection = jsonStringify(selection) === jsonStringify(previousSelection);

        if (isArray(snapSelection) && !isSameSelection) {
            snapSelection.sort(d3_ascending);
            previousSelection = snapSelection;
            brushInstance.setSelection(snapSelection, !brushed ? 0 : 250);
        }
    };
}

function _preventDefault() {
    return function(event) {
        //  ZC1927
        var sourceEvent = event.sourceEvent;
        if (sourceEvent && sourceEvent.preventDefault) {
            sourceEvent.preventDefault();
        }
    };
}

export const COLOR_BAND_BRUSH_DEFAULT_ACTION = {
    brush: filterRange,
    end: snapRange,
    start: _preventDefault
};
