// @zohocharts/zc/src/legend/colorband/colorband-brush-tooltip-content.js
// import { legend_helpers } from '../legend-helper';
import { legend_helpers } from '../../core/main-zc-init';
import { NUMERIC_DATATYPE, ORDINAL_DATATYPE } from '../../core/datatype-constants';
import { mathAbs, UNDEFINED } from '../../core/zc-constants';
import { defined, isArray, isUndefined } from '../../utils/zcutils';
import { getFormattedValue } from '../../formatter/formatter';
import { BOTTOM_ALIGN, TOP_ALIGN } from '../../core/align-constants';

/**
 *
 * @param brushComponent
 * @param i
 * @return {string}
 * @private
 */
export function _colorLegendBandBrushTooltipContent(brushComponent, i) {
    let config = brushComponent.helperOptions.spec.config,
        chartObj = brushComponent.chartObj,
        data = getActiveDomain(config, chartObj),
        // Formatter
        columnInfo = chartObj.dataset.getColumn('clr'),
        datatype = columnInfo.datatype,
        formatConf = columnInfo[datatype],
        options = {
            datatype: chartObj.dataset.getDataType('clr'),
            format: d => getFormattedValue(formatConf, d, datatype, chartObj, 'legend', UNDEFINED, '-'),
            ranges: config.colorBand.ranges,
            scale: chartObj.axes.clr.scale,
            reversed: legend_helpers.isReversedColorAxis(config),
            isHorizontal: isHorizontalOrient(brushComponent)
        };

    return getBrushTooltipContent(data, i, options);
}

export function getBrushTooltipContent(data, i, options) {
    let { datatype: dataType, format, ranges, scale, reversed, isHorizontal } = options;
    var labels = data,
        decimalCount,
        domainExtent;

    if (dataType === NUMERIC_DATATYPE) {
        if (!(ranges && ranges.length)) {
            var scale_domain = scale.domain();
            domainExtent = [scale_domain[0], scale_domain[scale_domain.length - 1]];
            var diff = mathAbs(domainExtent[1] - domainExtent[0]);

            if (diff) {
                //Test_779.json
                var hunds = Math.floor(Math.log(diff) / Math.LN10) - 2;
                decimalCount = Math.abs(Math.min(hunds, 0));
            }
        }
    }

    var index = i;
    if (options.encode !== 'z') {
        index = isHorizontal ? i : +!i;
    }

    if (reversed) {
        index = !index;
    }

    var val = labels[index ? labels.length - 1 : 0];
    // this.count = !this.count;
    if (defined(decimalCount) && domainExtent.indexOf(val) === -1) {
        //#ZC1283
        if (isArray(val)) {
            val = val.map(function(d) {
                return +d.toFixed(decimalCount);
            });
        } else {
            val = +val.toFixed(decimalCount);
        }
    }
    return format(val);
}

function isHorizontalOrient(brushObj) {
    var orient = brushObj.helperOptions.orient;
    return orient === TOP_ALIGN || orient === BOTTOM_ALIGN;
}

export function getActiveDomain(config, chartObj) {
    var colorbandOptions = config.colorBand,
        userRanges = colorbandOptions.ranges,
        isDiscrete = userRanges && userRanges.length,
        filterRange = config.filter.range,
        scale_domain = chartObj.axes.clr.scale.domain();

    if (isDiscrete) {
        var dataType = chartObj.dataset.getDataType('clr'),
            range = filterRange || (dataType === ORDINAL_DATATYPE ? scale_domain : userRanges);

        if (filterRange) {
            return getActiveDiscreteDomain(filterRange, userRanges);
        } else {
            return [range[0], range[range.length - 1]];
        }
    } else {
        return filterRange || scale_domain;
    }
}

export function getActiveDiscreteDomain(filterRange, userRanges) {
    var filterRange_copy = [filterRange[0], filterRange[1] + 0.000000000001],
        singleValueDomain = userRanges[0].length === 1,
        start,
        end;

    userRanges.forEach(function(d, i) {
        if (d[0] <= filterRange_copy[0]) {
            start = i;
        }
        if (d[d.length === 1 ? 0 : 1] <= filterRange_copy[1]) {
            end = i;
        }
    });

    isUndefined(start) ? (start = 0) : singleValueDomain ? start++ : null;
    isUndefined(end) ? (end = userRanges.length - 1) : null;
    return [userRanges[start], userRanges[end]];
}
