// @zohocharts/zc/src/legend/colorband/colorband-brush-helpers.js
import { d3_ascending, d3_extent, d3_merge } from '../../core/d3-constants';
import { d3_values, defined, validateNullVal } from '../../utils/zcutils';
import { mathMax, mathMin } from '../../core/zc-constants';
import { ORDINAL_DATATYPE } from '../../core/datatype-constants';

var MIN_VALUE = 0.000000000001;

/*
 *
 * @param handles
 * @param range
 */
export function styleHandles(options, handles, range) {
    handles.style('fill', function(d, i) {
        return getColor(range, i, options.isHorizontal, options.reversed, options.colorScale || options.scaleObj);
    });
}

/**
 *
 * @param options
 * @param selection
 * @return {*}
 */
export function getRange(options, selection) {
    let range = options.range, // [0, width] || [height: 0]
        ranges = options.ranges,
        userRanges_length = ranges && ranges.length;

    selection = getExtentWithinRange(selection, range);
    let newRange;
    if (userRanges_length) {
        newRange = getDomainForOrdinalScale(selection, options);
    } else {
        newRange = selection.map(d => options.scale.invert(d)).sort(d3_ascending);
    }
    return newRange;
}

/**
 *
 * @param options
 * @param selection
 * @return {*[]}
 */
export function snapHandlePosition(options, selection) {
    let snapSelection,
        ranges = options.ranges,
        userRanges_length = ranges && ranges.length,
        brushScale = options.scale;

    if (userRanges_length) {
        snapSelection = getExtentForOrdinalScale(getRange(options, selection), options);
    } else if (options.singleDomainContinuous) {
        snapSelection = options.singleDomainContinuous.map(d => brushScale(d));
    }
    return snapSelection;
}

/**
 *
 * @param extent
 * @param i
 * @param isHorizontal
 * @param reversed
 * @param scaleObj
 * @return {*}
 */
function getColor(extent, i, isHorizontal, reversed, scaleObj) {
    var colorIndex = isHorizontal ? [0, extent.length - 1] : [extent.length - 1, 0];

    if (reversed) {
        colorIndex.reverse();
    }

    var clrValue = extent[colorIndex[i]];
    return scaleObj(clrValue);
}

/**
 *
 * @param extent
 * @param brushScale
 * @param scale_domain
 * @param userRanges
 * @param datatype
 * @param isHorizontal
 * @param reversed
 * @return {*[]}
 */
function getDomainForOrdinalScale(extent, options) {
    let { scale_domain, datatype, isHorizontal, reversed, scale: brushScale, ranges: userRanges } = options,
        discrete_ranges = brushScale.domain().map(brushScale),
        brushScale_range = d3_values(discrete_ranges).sort(d3_ascending),
        brushScale_domain = brushScale.domain(),
        rangeBand = brushScale.bandwidth(),
        startIndex,
        endIndex;

    // Find start and end index
    brushScale_range.some(function(d, i) {
        if (extent[0] >= d && extent[0] <= d + rangeBand) {
            var diff = extent[0] - d;
            if (diff <= 0.5 && diff >= 0) {
                // antialiasing artifacts
                startIndex = mathMax(i - 1, 0);
            } else {
                startIndex = i;
            }
        }
        if (extent[1] >= d && extent[1] <= d + rangeBand) {
            var diff = d + rangeBand - extent[1];
            if (diff <= 0.5 && diff >= 0) {
                // antialiasing artifacts
                endIndex = mathMin(i + 1, brushScale_range.length - 1);
            } else {
                endIndex = i;
            }
        }
        return defined(startIndex) && defined(endIndex);
    });

    startIndex = mathMax(0, validateNullVal(startIndex, 0));
    endIndex = mathMax(0, validateNullVal(endIndex, brushScale_domain.length - 1)); //#ZC1319

    var domain = brushScale_domain;
    if (datatype !== ORDINAL_DATATYPE && userRanges[0].length === 1) {
        // If ranges are single element like [50, 60, 70] (e.g: http://localhost/#/demoCharts/heatmap5), need to make it as range --> [[min, 50], [50, 60], [60, 70]]
        domain = getPairs(scale_domain, domain);
    }

    if (options.encode !== 'z' ? (!isHorizontal && !reversed) || (isHorizontal && reversed) : reversed) {
        var temp = startIndex;
        startIndex = brushScale_range.length - 1 - endIndex;
        endIndex = brushScale_range.length - 1 - temp;
    }

    domain = domain.slice(startIndex, endIndex + 1, 0);
    if (datatype === ORDINAL_DATATYPE) {
        extent = domain;
    } else {
        extent = d3_extent(d3_merge(domain));
        extent = updateFilterRange(extent, scale_domain);
    }
    return extent;
}

/**
 *
 * @param filterRange
 * @param brushScale
 * @param scale_domain
 * @param userRanges
 * @param datatype
 * @return {*[]}
 */
export function getExtentForOrdinalScale(filterRange, options) {
    var { scale: brushScale, ranges: userRanges, scale_domain, datatype } = options,
        rangeBandBy2 = brushScale.bandwidth() / 2,
        extent;

    if (datatype === ORDINAL_DATATYPE) {
        extent = [brushScale(filterRange[0]) + rangeBandBy2, brushScale(filterRange[filterRange.length - 1]) + rangeBandBy2];
    } else {
        var brushScale_domain = brushScale.domain(),
            //start and end index of userRange
            startIndex,
            endIndex; //start and end index of userRange
        filterRange = updateFilterRange(filterRange, scale_domain);

        brushScale_domain.forEach(function(d, i) {
            if (filterRange[0] >= d[0]) {
                startIndex = userRanges[0].length === 1 ? i + 1 : i;
            }
            if (!defined(endIndex) && filterRange[1] <= d[0]) {
                endIndex = userRanges[0].length === 1 ? i : i - 1;
            }
        });

        startIndex = mathMax(0, validateNullVal(startIndex, 0));
        endIndex = mathMax(0, validateNullVal(endIndex, brushScale_domain.length - 1));

        extent = [brushScale(brushScale_domain[startIndex]) + rangeBandBy2, brushScale(brushScale_domain[endIndex]) + rangeBandBy2];
    }

    if (extent[0] === extent[1]) {
        extent[0] -= MIN_VALUE;
        extent[1] += MIN_VALUE;
    }
    return extent;
}

/**
 *
 * @param extent
 * @param range
 * @return {*}
 */
function getExtentWithinRange(extent, range) {
    //Refer: legend filter issue
    if (range[1] > range[0]) {
        extent[1] = mathMin(extent[1], range[1]);
        extent[0] = mathMax(extent[0], range[0]);
    } else {
        extent[1] = mathMax(extent[1], range[1]);
        extent[0] = mathMin(extent[0], range[0]);
    }
    return extent;
}

function updateFilterRange(filterRange, scale_domain) {
    // If [[10, 20], [20, 30]] is the user ranges (20 is present twice), the filterRange is should be modified
    // Refer legend-brush-1

    if (!isDistinct(scale_domain, filterRange[1])) {
        filterRange[1] -= MIN_VALUE;
    }
    return filterRange;
}

function isDistinct(array, value) {
    return array.indexOf(value) === array.lastIndexOf(value);
}

function getPairs(scale_domain, domain) {
    var min = scale_domain[2],
        max = scale_domain[scale_domain.length - 1 - 2],
        pairs = [];
    for (var i = 0; i < domain.length; i++) {
        pairs.push([min, domain[i][0]]);
        min = domain[i][0];
    }
    pairs.push(min, max);
    return pairs;
}
