// @zohocharts/zc/src/treemap/canvasTreemapchart.js
//$Id$
//canvasTreemapchart.js
import { Registry } from '../core/zc-registry';
import { TreemapHighlighter } from './treemap-highlight';
import { treemapLayout } from './treemap-layout';
import { colorscale_helpers } from '../axis/colorscale-helpers';
import { isMultiColoring, isWithinColorRange, getRawColor, rgba_canvas } from '../utils/colorutils';
import { NULL, TRANSPARENT, NONE } from '../core/zc-constants';
import { canvasUtils } from '../utils/canvasutils';
// import { legend_helpers } from '../legend/legend-helper';
import { isUndefined, defined, getBorderRadius } from '../utils/zcutils';
import { effects } from '../effects/zceffects';
import { PERCENT_SHOWAS } from '../label/showas-constants';
import { labelRenderer, legend_helpers, canvas_plot_renderer } from '../core/main-zc-init';
import { treemapUtils } from './treemap-utils';
import { isFacet } from '../axis/hierarchy-axis-utils';

let annotation = Registry.getComponent('annotation');
canvas_plot_renderer.treemap = function() {
    var commonRendererProp,
        getLevel = treemapUtils.getLevel,
        getSeriesIndex = treemapUtils.getSeriesIndex,
        getItemIndex = treemapUtils.getItemIndex,
        layoutObj;

    function chart(context, seriesdata, effect, filterIndex, encode, _index) {
        var rp = commonRendererProp.params,
            chartObj = rp.chartObj,
            commonPlotOptions = rp.commonPlotOptions,
            notesArray = rp.notesArray,
            labelQueue = rp.labelQueue,
            { rendererIndex } = rp,
            isSelectionCanvas = context.canvas.className == 'selection-canvas' ? true : false,
            selectionArray = chartObj.cache.clickSelectionData;

        chart.layout = layoutObj = isHighlight
            ? layoutObj
            : new treemapLayout(commonRendererProp, isFacet(chartObj) || chartObj.isAxisedHierarchyChart ? seriesdata : chartObj.seriesdata);
        var hasColorScale = colorscale_helpers.isQuantileColorScale(chartObj),
            multiColoring = isMultiColoring(commonPlotOptions, chartObj),
            borderRadius = getBorderRadius(commonPlotOptions.borderRadius),
            parentOptions = commonPlotOptions.parent,
            dimension = layoutObj.dimension,
            leafNodes = layoutObj.leafNodes,
            seriesWiseNodes = layoutObj.seriesWiseNodes,
            processedData = layoutObj.processedData,
            fillColor = layoutObj.fillColor,
            isSelectionContainer = !(chartObj.container.select('.selection-canvas').empty()),
            isHighlight = context.canvas.className === 'highlight-canvas',
            selectionArray = chartObj.cache.clickSelectionData;

        context.save();
        context.translate(dimension.x, dimension.y);
        chart.total = processedData[0].value;

        if (isHighlight) {
            var index = _index;
            processedData = processedData.filter(function(d) {
                return getSeriesIndex(d) === index && !(hasColorScale && d.children);
            });

            if (hasColorScale) {
                processedData = legend_helpers.filterSelectionByColor(processedData, true, filterIndex, chartObj);
            }
        }
        if(isSelectionContainer ){
            if(isSelectionContainer && defined(selectionArray[0])){
                chartObj.container.select('.zc-canvas').style("visibility", "hidden");
            }
            else{
                chartObj.container.select('.selection-canvas').style("visibility", "hidden");
                chartObj.container.select('.zc-canvas').style("visibility", "visible");
            }
        }
        
        processedData.forEach(function(d, i) {
            var seriesIndex = getSeriesIndex(d),
                itemIndex = getItemIndex(d),
                tileBound = treemapUtils.getTileBound(layoutObj, d),
                x = tileBound.x,
                y = tileBound.y,
                w = tileBound.width,
                h = tileBound.height,
                customStrokeColor = (d.children ? parentOptions.strokeColor : NULL) || commonPlotOptions.strokeColor,
                strokeClr;

            if (!d.depth) {
                strokeClr = TRANSPARENT;
            } else if (
                isHighlight &&
                isUndefined(effect) &&
                chartObj.legendHighlight.getLegendHighlightEffect(index, 'selectedSeries', d.data.paneid) !== NONE
            ) {
                strokeClr = 'black';
            } else if (customStrokeColor) {
                strokeClr = customStrokeColor;
            } else if (d.children) {
                strokeClr = d.data.arcColor;
            }

            var drawCell = function(context, fillStyle) {
                context.save();
                context.translate(x + w / 2, y + h / 2);
                context.beginPath();
                context.lineWidth = (d.children ? parentOptions.strokeWidth : NULL) || commonPlotOptions.strokeWidth;
                context.strokeStyle = rgba_canvas(
                    strokeClr,
                    (d.children ? parentOptions.strokeOpacity : NULL) || commonPlotOptions.strokeOpacity
                );
                context.fillStyle = fillStyle;

                canvasUtils.roundedRect(context, -w / 2, -h / 2, w, h, borderRadius);
                context.fill();
                context.stroke();
                context.closePath();
                context.restore();
            };

            if (chartObj.imagePallete && isWithinColorRange(chartObj.dataset.getClr(d), chartObj) && d.depth) {
                effects.canvas_patterns.imageFill(
                    { x: x, y: y, context: context, name: getRawColor(chartObj, multiColoring ? itemIndex : seriesIndex, d, seriesIndex) },
                    drawCell
                );
            } else {
                var fillStyle;
                if (defined(effect) && !isSelectionCanvas){
                    fillStyle = effect;
                }  
                else if(isSelectionCanvas && isUndefined(d.children)){
                    if(defined(selectionArray[0]) && !d.data['clickSelection']){
                        fillStyle = effect;
                    }
                    else{
                        fillStyle = rgba_canvas(d.data.arcColor, parentOptions.fillOpacity);
                    }
                }
                else if (!d.children) {
                    fillStyle = fillColor[seriesIndex](seriesIndex, d.data, NULL, NULL, commonPlotOptions.fillOpacity, [w, h]);
                }  
                else {
                    fillStyle = rgba_canvas(d.data.arcColor, parentOptions.fillOpacity);
                }
                drawCell(context, fillStyle);
            }
        });

        context.restore();

        if (labelRenderer) {
            var transform = [dimension.x, dimension.y];
            var totalFunc = function(d, showAs) {
                if (showAs === PERCENT_SHOWAS) {
                    return chart.total;
                } else {
                    var node = chart.getProcessedNode(d);
                    return node.parent.value;
                }
            };
            seriesdata.forEach(function(d, i) {
                let _currentRendererIndex = isFacet(chartObj) ? rendererIndex + i : i,
                    paneid = isFacet(chartObj) ? `${d.row}_${d.column}` : i;

                labelQueue.push({
                    renderer: chart,
                    charttype: 'hierarchy',
                    chartObj: chartObj,
                    rendererIndex: _currentRendererIndex,
                    paneid,
                    args: [seriesWiseNodes[i], i, NULL, NULL, totalFunc, leafNodes[i], transform]
                });

                labelQueue.push({
                    renderer: chart,
                    charttype: 'hierarchy',
                    chartObj: chartObj,
                    rendererIndex: _currentRendererIndex,
                    paneid,
                    args: [seriesWiseNodes[i], i, NULL, NULL, totalFunc, [seriesWiseNodes[i][0]], transform, true]
                });
            });
        }

        //Annotation preprocessing
        if (!annotation._empty && !isHighlight) {
            chartObj.seriesdata.forEach(function(series, i) {
                notesArray = annotation.preprocessAnnotations(
                    chartObj.seriesdata,
                    i,
                    function(dt, k, j) {
                        var node = chart.getProcessedNode(dt);
                        var tileBound = treemapUtils.getTileBound(layoutObj, node);
                        return [tileBound.x + dimension.x, tileBound.y + dimension.y];
                    },
                    notesArray,
                    chartObj,
                    rp
                );
            });
        }

        chart.chartRendered = true;
        return chart;
    }

    chart.commonRendererProp = function(_) {
        if (!arguments.length) {
            return commonRendererProp;
        }
        commonRendererProp = _;
        return chart;
    };

    chart.getProcessedNode = function(d) {
        return layoutObj.nodes[getLevel(d)];
    };

    chart.getHighlighter = function(chartObj, index) {
        return (chart.highlighter = new TreemapHighlighter(chartObj, chart));
    };
    return chart;
};

export var _canvasTreemapChart = canvas_plot_renderer.treemap;
