// @zohocharts/zc/src/treemap/treemap-layout.js
//$Id$
//treemap-layout.js
import { rendererBoundUtils } from '../plot/renderer-bound-utils';
import { d3_scaleLinear, d3_tiles, d3_treemap, d3_hsl, d3_hierarchy } from '../core/d3-constants';
import { parseShortHandValue, isArray, isProperObject, isNumber, isString, getPropVal } from '../utils/zcutils';
import { DataProcessor } from '../core/data-processor';
import { NULL, mathAbs } from '../core/zc-constants';
import { getColorCopy, isMultiColoring, getFillColor, getStrokeColor, getRawColor } from '../utils/colorutils';
import { getChildrenCount } from '../core/hierarachy-data-processor';
import { treemapUtils } from './treemap-utils';
import { isFacet } from '../axis/hierarchy-axis-utils';

export var treemapLayout = function(commonRendererProp, data) {
    var self = this;
    self.data = data;
    var rp = (self.rp = commonRendererProp.params),
        { bound } = rp;

    self.chartObj = rp.chartObj;
    self.commonRendererProp = commonRendererProp;
    self.dimension = rendererBoundUtils.getAvailableBound(bound, rp.commonPlotOptions);

    self.createScale();
    self.layout = self.getLayout();
    self.processData();
};

treemapLayout.prototype.createScale = function() {
    var dimension = this.dimension;
    //x scale
    let xScale = d3_scaleLinear()
        .domain([0, dimension.width])
        .range([0, dimension.width]);

    // y sclae
    let yScale = d3_scaleLinear()
        .domain([0, dimension.height])
        .range([0, dimension.height]);

    this.xScale = xScale;
    this.yScale = yScale;
};

treemapLayout.prototype.getLayout = function() {
    var self = this,
        dimension = self.dimension,
        commonPlotOptions = this.rp.commonPlotOptions,
        parentOptions = commonPlotOptions.parent,
        paddingOptions = commonPlotOptions.padding,
        mode = commonPlotOptions.mode;

    paddingOptions = isProperObject(paddingOptions) ? paddingOptions : parseShortHandValue(paddingOptions);

    let padding = parseShortHandValue();  // #ZC4723 - callback support for treemap.padding
    for (const key in paddingOptions) {
        const _padding = paddingOptions[key];
        padding[key] = isNumber(_padding) ? function (d) { return d.depth && d.value ? _padding : 0 }
            : isString(_padding) ? getPropVal(window, _padding) : _padding;
    }

    return d3_treemap()
        .size([dimension.width, dimension.height])
        .tile(d3_tiles(mode))
        .round(true)
        .paddingInner(function(d) {
            //padding between children
            if (d.value) {
                return (d.children[0] && d.children[0].children ? parentOptions : commonPlotOptions).interPadding || 0;
            } else {
                return 0;
            }
        })
        .paddingTop(padding.top)
        .paddingBottom(padding.bottom)
        .paddingLeft(padding.left)
        .paddingRight(padding.right);
};

treemapLayout.prototype.getValue = function(d) {
    var self = this,
        chartObj = self.chartObj,
        rp = self.rp,
        curSeriesData = getCurrentSeriesData(self, d),
        yaxiscolumnorder = DataProcessor.helpers.getParsedYAxisOrder(curSeriesData.yaxiscolumnorder),
        y = chartObj.dataset.getY(d, yaxiscolumnorder) || 0,
        negativeAsAbsolute = rp.commonPlotOptions.negativeAs === 'absolute';

    if (treemapUtils.isDisabled(curSeriesData, d)) {
        return 0;
    }

    if (y < 0) {
        return negativeAsAbsolute ? mathAbs(y) : 0;
    }
    return y;
};

treemapLayout.prototype.getChildren = function(node) {
    var level = treemapUtils.getLevel(node),
        depth = level && level.split('_').length;

    if (!this.chartObj.dataObject.hasHierarchyData && depth == 1 && node.length < 2 && node[0] && isArray(node[0][0])) {
        return node[0];
    } else if (isArray(node[0])) {
        return node;
    } else {
        return NULL;
    }
};

treemapLayout.prototype.processData = function() {
    var self = this,
        data = self.data,
        chartObj = self.chartObj,
        commonRendererProp = self.commonRendererProp,
        dimension = self.dimension,
        rp = self.rp,
        commonPlotOptions = rp.commonPlotOptions,
        color = rp.color,
        { rendererIndex = 0 } = rp,
        colorScheme = commonPlotOptions.colorScheme || 'solid', //solid | tints | shades
        //previous sibling info for fadeOneByOne animation
        previousSiblingsCounter = 0,
        previousSiblingsCount = {},
        colorLighter = [],
        childColorIndex = [],
        seriesWiseNodes = [],
        leafNodes = [],
        nodes = [],
        fillColor = [],
        strokeColor = [],
        root = d3_hierarchy(data, function(d) {
            return self.getChildren(d);
        }).sum(function(d) {
            return self.getValue(d);
        }),
        // .sort(d3_ascending);
        processedData = self.layout(root).descendants(),
        paneid = `${data[0].row}_${data[0].column}`;

    data.forEach(function(d, i) {
        let rendererConf = chartObj.rendererConf[rendererIndex + i];
        var nonRGBAColor = getRawColor(chartObj, i, null, i, paneid);

        fillColor[i] = getFillColor(chartObj, i, paneid, [dimension.width, dimension.height], null, null, null, rendererConf.params);
        strokeColor[i] = getStrokeColor(chartObj, i, paneid, null, rendererConf.params);

        childColorIndex[i] = 0;
        seriesWiseNodes[i] = [];
        seriesWiseNodes[i].disabled = d.disabled;
        seriesWiseNodes[i].plotoptions = d.plotoptions;
        colorLighter[i] = d3_scaleLinear()
            .domain([0, getChildrenCount(d)])
            .range([
                nonRGBAColor,
                getColorCopy(nonRGBAColor, colorScheme === 'solid' ? 0 : colorScheme === 'shades' ? -1 : 1, 'brighter', d3_hsl)
            ]);
    });

    processedData.forEach(function(d, i) {
        var level = treemapUtils.getLevel(d),
            seriesIndex = treemapUtils.getSeriesIndex(d),
            itemIndex = treemapUtils.getItemIndex(d),
            curSeriesData = getCurrentSeriesData(self, d);

        d.index = i;
        d.data.paneid = paneid;
        if (d.depth) {
            if (d.data.color) {
                d.data.arcColor = d.data.color;
            } else {
                var c = isMultiColoring(commonPlotOptions, chartObj)
                    ? color[itemIndex % color.length]
                    : colorLighter[seriesIndex](childColorIndex[seriesIndex]);
                d.data.arcColor = c;
            }
            nodes[level] = d;
        }

        if (!d.children) {
            if (!leafNodes[seriesIndex]) {
                leafNodes[seriesIndex] = [];
            }
            leafNodes[seriesIndex].push(d);
            childColorIndex[seriesIndex]++;
        } else {
            d.data.arcColor = treemapUtils.getParentColor(commonRendererProp, d, paneid);
        }

        if (d.depth) {
            seriesWiseNodes[seriesIndex].push(d);
        }

        if (!treemapUtils.isDisabled(curSeriesData, d)) {
            previousSiblingsCount[level] = previousSiblingsCounter++;
        }
    });

    chartObj.renderer[0].fillColorFunc = fillColor[0];

    self.processedData = processedData;
    self.previousSiblingsCount = previousSiblingsCount;
    self.seriesWiseNodes = seriesWiseNodes;
    self.leafNodes = leafNodes;
    self.nodes = nodes;
    self.fillColor = fillColor;
    self.strokeColor = strokeColor;
};

function getCurrentSeriesData(layoutObj, d) {
    var data = layoutObj.data,
        chartObj = layoutObj.chartObj,
        seriesIndex = treemapUtils.getSeriesIndex(d) || 0;

    return isFacet(chartObj) ? data[seriesIndex].node : chartObj.seriesdata[seriesIndex];
}
