// @zohocharts/zc/src/treemap/treemap-highlight.js
import { BarHighlighter } from '../bar/bar-highlight';
import { TRANSPARENT } from '../core/zc-constants';
import { HIGHLIGHT_CONTEXT } from '../events/visualEffects/highlight/highlight-context';
import { createElement, roundedRect, checkRadiusWithinBoundary } from '../utils/zcsvgutils';
import { inherit, getBorderRadius } from '../utils/zcutils';
import { euclideanDistance } from '../utils/geometry/_point';
import { treemapUtils } from './treemap-utils';
import { isMultiColoring, getRawColor, getFillColor } from '../utils/colorutils';
import { isPointVisible } from '../events/events-utils';
import { getRendererIndex } from '../core/data-parser';

/*
 Doubts:
 1) Whether sharedseries will be applicable?
 Pending:
 1) Hierarchy data. (e.g): treemap-01
 2) scroll treemap
 */
export function TreemapHighlighter(chart, renderer) {
    this.chart = chart;
    this.renderer = renderer;
}

inherit(TreemapHighlighter, BarHighlighter);
let proto = TreemapHighlighter.prototype;

proto.renderingContext = function() {
    return HIGHLIGHT_CONTEXT.get('nonaxis').call(this, this.chart);
};

proto.translateContext = function(point) {
    let chart = this.chart,
        renderer = getRenderer(point, chart, this.renderer),
        dimension = renderer.layout.dimension,
        bound = renderer.commonRendererProp().bound(),
        plotArea = chart.plotarea;

    return [plotArea.left + dimension.x + bound.x, plotArea.top + dimension.y + bound.y];
};

proto.data = function(point) {
    let chart = this.chart,
        renderer = getRenderer(point, chart, this.renderer),
        layout = renderer.layout,
        dimension = layout.dimension,
        bound = renderer.commonRendererProp().bound();

    return function() {
        if (!chart.cache.data[point._key]) {
            chart.cache.data[point._key] = {};
        }

        let cache = chart.cache.data[point._key];
        let node = renderer.getProcessedNode(point);
        let tileBound = treemapUtils.getTileBound(layout, node);

        cache._x = point._x = cache._x || tileBound.x + dimension.x + bound.x + tileBound.width / 2;
        cache._y = point._y = cache._y || tileBound.y + dimension.y + bound.y + tileBound.height / 2;
        cache._width = point._width = cache._width || tileBound.width;
        cache._height = point._height = cache._height || tileBound.height;
        point._centric = true;
        cache._size = point._size =
            2 * euclideanDistance({ x: point._x, y: point._y }, { x: point._x - tileBound.width / 2, y: point._y - tileBound.height / 2 });
        cache._uid = point._uid = chart.dataset.getX(point);
        cache._visible = point._visible = isPointVisible(chart, point);
        cache._point = point;

        return point;
    };
};

proto.options = function(point) {
    let chart = this.chart,
        renderer = getRenderer(point, chart, this.renderer),
        layout = renderer.layout,
        rp = renderer.commonRendererProp().params,
        dimension = layout.dimension,
        { seriesIndex } = point,
        tileBound = getTileBound(renderer, point),
        borderRadius = getBorderRadius(rp.commonPlotOptions.borderRadius),
        multiColoring = isMultiColoring(rp.commonPlotOptions, chart),
        rawColor = getRawColor(chart, multiColoring ? point.itemIndex : seriesIndex, point, seriesIndex);

    if (chart.plot.renderer.mode === 'canvas') {
        //#ZC1193
        borderRadius = checkRadiusWithinBoundary(borderRadius, tileBound.width, tileBound.height, true);
    }

    return {
        radius: borderRadius,
        color: rawColor,
        strokeColor: TRANSPARENT,
        fillColor: (function() {
            let func = getFillColor(chart, seriesIndex, null, [dimension.width, dimension.height], null, 'svg');
            return func(seriesIndex, point, null, null, null, [tileBound.width, tileBound.height]);
        })()
    };
};

proto.path = function(point, model) {
    let self = this,
        context = model.context.call(self),
        translate = model.translate.call(self, point),
        chart = this.chart,
        renderer = getRenderer(point, chart, this.renderer),
        rp = renderer.commonRendererProp().params,
        //
        tileBound = getTileBound(renderer, point),
        path = roundedRect(tileBound.x, tileBound.y, tileBound.width, tileBound.height, getBorderRadius(rp.commonPlotOptions.borderRadius));

    context.style('display', '').style('opacity', null);

    let groupContext = createElement('g', `.treemap-highlight${point.paneid}${point.seriesIndex}${point.subSeriesIndex}`, [point], context);
    groupContext.attr('transform', `translate(${translate[0]},${translate[1]})`);

    let pathContext = createElement('path', `.treemap-highlight`, [point], groupContext).attr('d', path);
    return pathContext;
};

function getRenderer(point, chart, renderer) {
    return point ? chart.renderer[getRendererIndex(point, chart)] : renderer;
}

function getTileBound(renderer, point) {
    let node = renderer.getProcessedNode(point),
        layout = renderer.layout;
    return treemapUtils.getTileBound(layout, node);
}
