// @zohocharts/zc/src/treemap/treemapchart.js
//$Id$
//treemapchart.js
import { Registry } from '../core/zc-registry';
import { svg_plot_renderer } from '../core/main-zc-init';
import { treemapUtils } from './treemap-utils';
import { treemapLayout } from './treemap-layout';
import { getAnimationConf } from '../utils/animationutils';
import { d3_ease_linear, d3_select } from '../core/d3-constants';
import {
    FADEALL_ANIMATION,
    FADEONEBYONE_ANIMATION,
    FADERANDOM_ANIMATION,
    RANDOMOUT_ANIMATION,
    HORIZONTALALL_ANIMATION,
    VERTICALALL_ANIMATION
} from '../core/animation-constants';
import { appendEle, roundedRect } from '../utils/zcsvgutils';
import { NULL, TRANSPARENT, POINTER } from '../core/zc-constants';
import { PATH_ELEMENT } from '../core/element-constants';
import { PERCENT_SHOWAS } from '../label/showas-constants';
import { defined, getBorderRadius } from '../utils/zcutils';
import { labelRenderer } from '../core/main-zc-init';
import { mathRandom } from '../core/zc-constants';
import { isFacet } from '../axis/hierarchy-axis-utils';
import { setOpacityForTickFilterElements, setPivotAttributes } from '../plot/svg/set-pivot-attr';
import { TreemapHighlighter } from './treemap-highlight';

let annotation = Registry.getComponent('annotation');
/**
 *
 * @returns {chart} treemap renderer.
 */

/* PENDING:
 * Annotation for hierarchy data
 */
svg_plot_renderer.treemap = function() {
    var isredraw = false,
        commonRendererProp,
        treemapGroup,
        path,
        chartObj,
        commonPlotOptions,
        // clickedItem,
        // xScale,
        // yScale,

        getLevel = treemapUtils.getLevel,
        getSeriesIndex = treemapUtils.getSeriesIndex,
        // getItemIndex = treemapUtils.getItemIndex,

        layoutObj,
        dimension,
        nodes,
        leafNodes,
        seriesWiseNodes;

    function chart(selection) {
        selection.each(function(data) {
            //renderer config values
            var rp = commonRendererProp.params;
            chartObj = rp.chartObj;
            commonPlotOptions = rp.commonPlotOptions;
            var notesArray = rp.notesArray,
                systemconf = chartObj.systemConf,
                labelQueue = rp.labelQueue,
                { rendererIndex } = rp;
            if(chartObj.customVisual){
                chart.chartRendered = true;
                return;
            }
            chart.layout = layoutObj = new treemapLayout(commonRendererProp, data);
            dimension = layoutObj.dimension;
            nodes = layoutObj.nodes;
            leafNodes = layoutObj.leafNodes;
            seriesWiseNodes = layoutObj.seriesWiseNodes;

            var strokeWidth = commonPlotOptions.strokeWidth,
                strokeOpacity = commonPlotOptions.strokeOpacity,
                fillOpacity = commonPlotOptions.fillOpacity,
                fillColor = layoutObj.fillColor,
                strokeColor = layoutObj.strokeColor,
                borderRadius = getBorderRadius(commonPlotOptions.borderRadius),
                transform = [dimension.x, dimension.y];

            //Animation variables
            var animationOptions = getAnimationConf(commonPlotOptions, NULL, chartObj, 'treemap');
            var isAnimationEnable = animationOptions.enabled,
                animationType = animationOptions.type,
                easingType = isredraw ? d3_ease_linear : animationOptions.easingType,
                transitionDuration = isredraw ? (isAnimationEnable ? 200 : 0) : animationOptions.duration,
                dataCount = chartObj.dataObject.totalnoofdatas,
                duration = isredraw
                    ? transitionDuration
                    : animationType === FADEONEBYONE_ANIMATION
                    ? transitionDuration / dataCount
                    : transitionDuration,
                delay = isredraw
                    ? 0
                    : animationType === FADEONEBYONE_ANIMATION
                    ? function(d) {
                          return layoutObj.previousSiblingsCount[getLevel(d)] * (transitionDuration / dataCount);
                      }
                    : animationType === FADERANDOM_ANIMATION
                    ? function() {
                          return mathRandom() * dataCount * (transitionDuration / dataCount);
                      }
                    : 0,
                initialOpacity =
                    animationType === FADEALL_ANIMATION ||
                    animationType === FADERANDOM_ANIMATION ||
                    animationType === FADEONEBYONE_ANIMATION
                        ? 0
                        : NULL;

            var cursor = commonPlotOptions.events
                    ? commonPlotOptions.events.cursor || POINTER
                    : systemconf.chart.plot.events.cursor || POINTER,
                parentOptions = commonPlotOptions.parent,
                parentCursor = parentOptions.events.cursor;

            treemapGroup = appendEle(selection, 'g', [data], '.', 'class', 'hierarchychart');
            treemapGroup.attr('transform', `translate(${dimension.x},${dimension.y})`).styles({
                'stroke-width': strokeWidth,
                'stroke-opacity': strokeOpacity,
                'fill-opacity': fillOpacity,
                cursor: cursor
            });

            var processedData = layoutObj.processedData;
            chart.total = processedData[0].value;
            path = appendEle(treemapGroup, PATH_ELEMENT, processedData);
            if (!chart.total) {
                path.remove();
            }

            path.styles({
                fill: function(d) {
                    if (!d.children) {
                        var tileBound = treemapUtils.getTileBound(layoutObj, d),
                            seriesIndex = getSeriesIndex(d);

                        return (d.fillColor = fillColor[seriesIndex](seriesIndex, d.data, NULL, NULL, NULL, [
                            tileBound.width,
                            tileBound.height
                        ]));
                    } else {
                        return (d.fillColor = d.data.arcColor);
                    }
                },
                stroke: function(d) {
                    var customStrokeColor = (d.children ? parentOptions.strokeColor : NULL) || commonPlotOptions.strokeColor,
                        seriesIndex = getSeriesIndex(d);

                    if (!d.depth) {
                        return TRANSPARENT;
                    } else if (customStrokeColor) {
                        return customStrokeColor;
                    } else if (d.children) {
                        return d.data.arcColor;
                    } else {
                        return strokeColor[seriesIndex](seriesIndex, d.data);
                    }
                },
                visibility: function(d) {
                    return d.value ? NULL : 'hidden';
                }
            })
                .attr('class', d => {
                    if (defined(getLevel(d))) {
                        return `zc_treemap series_${getSeriesIndex(d)}${d.children ? ' zchierarchyparent' : ''}`;
                    }
                    return 'zc_treemap';
                })
                .attr('id', d => `tree_${getLevel(d)}`);

            var parentNodes = treemapGroup.selectAll(`${PATH_ELEMENT}.zchierarchyparent`);
            parentNodes.styles({
                'stroke-width': parentOptions.strokeWidth,
                'stroke-opacity': parentOptions.strokeOpacity,
                'fill-opacity': parentOptions.fillOpacity,
                cursor: parentCursor
            });

            if (!isredraw) {
                path.attr('transform', function(d) {
                    var tileBound = treemapUtils.getTileBound(layoutObj, d),
                        tX = tileBound.x + tileBound.width / 2,
                        tY = tileBound.y + tileBound.height / 2;

                    if (animationType === RANDOMOUT_ANIMATION) {
                        tX = dimension.width / 2;
                        tY = dimension.height / 2;
                    } else if (animationType === HORIZONTALALL_ANIMATION) {
                        tX = tileBound.width / 2;
                    } else if (animationType === VERTICALALL_ANIMATION) {
                        tY = tileBound.height / 2;
                    }

                    return `translate(${tX},${tY})`;
                })
                    .attr('d', function(d) {
                        var tileBound = treemapUtils.getTileBound(layoutObj, d),
                            x = -tileBound.width / 2,
                            y = -tileBound.height / 2,
                            w = tileBound.width,
                            h = tileBound.height;

                        if (animationType === HORIZONTALALL_ANIMATION) {
                            w = 0;
                        } else if (animationType === VERTICALALL_ANIMATION) {
                            h = 0;
                        }

                        return roundedRect(x, y, w, h, borderRadius);
                    })
                    .style('opacity', initialOpacity);
            }

            setPivotAttributes(path, chartObj, 'treemap');
            setOpacityForTickFilterElements(path, chartObj, 'treemap');

            path.transition()
                .duration(duration)
                .delay(delay)
                .ease(easingType)
                .attr('transform', function(d) {
                    var tileBound = treemapUtils.getTileBound(layoutObj, d);
                    return `translate(${tileBound.x + tileBound.width / 2},${tileBound.y + tileBound.height / 2})`;
                })
                .attr('d', function(d) {
                    var tileBound = treemapUtils.getTileBound(layoutObj, d);
                    return roundedRect(-tileBound.width / 2, -tileBound.height / 2, tileBound.width, tileBound.height, borderRadius);
                })
                .style(
                    'opacity',
                    isFacet(chartObj)
                        ? function() {
                              return d3_select(this).attr('filtered-opacity') || 1;
                          }
                        : NULL
                );

            chartObj.timerObj[chartObj.timerObj.length] = chart.renderedTimer = setTimeout(function() {
                if (labelRenderer) {
                    var totalFunc = function(d, showAs) {
                        if (showAs === PERCENT_SHOWAS) {
                            return chart.total;
                        } else {
                            var node = chart.getProcessedNode(d);
                            return node.parent.value;
                        }
                    };

                    data.forEach(function(d, i) {
                        let _currentRenderer = chartObj.renderer[isFacet(chartObj) || chartObj.isAxisedHierarchyChart ? rendererIndex + i : 0],
                            _currentRendererIndex = isFacet(chartObj) ? rendererIndex + i : i,
                            paneid = isFacet(chartObj) ? `${d.row}_${d.column}` : i;

                        labelQueue.push({
                            renderer: _currentRenderer,
                            charttype: 'hierarchy',
                            chartObj: chartObj,
                            rendererIndex: _currentRendererIndex,
                            paneid,
                            args: [seriesWiseNodes[i], i, NULL, NULL, totalFunc, leafNodes[i], transform]
                        });

                        labelQueue.push({
                            renderer: _currentRenderer,
                            charttype: 'hierarchy',
                            chartObj: chartObj,
                            rendererIndex: _currentRendererIndex,
                            paneid,
                            args: [seriesWiseNodes[i], i, NULL, NULL, totalFunc, [seriesWiseNodes[i][0]], transform, true]
                        });
                    });
                }

                //Annotation preprocessing
                if (!annotation._empty) {
                    data.forEach(function(series, i) {
                        notesArray = annotation.preprocessAnnotations(
                            data,
                            i,
                            function(dt, k, j) {
                                var node = chart.getProcessedNode(dt);
                                var tileBound = treemapUtils.getTileBound(layoutObj, node);
                                return [tileBound.x + dimension.x, tileBound.y + dimension.y];
                            },
                            notesArray,
                            chartObj,
                            rp
                        );
                    });
                }

                isredraw = true;
                chart.chartRendered = true;
            }, transitionDuration + 17 + (delay ? dataCount * 4 : 0));
        });
        return chart;
    }
    chart.commonRendererProp = function(_) {
        if (!arguments.length) {
            return commonRendererProp;
        }
        commonRendererProp = _;
        return chart;
    };
    // chart.defaultClickFunc = function(d) {
    // 	var rp = commonRendererProp.params;
    // 	if (clickedItem === getLevel(d)) {
    // 		xScale.domain([0, dimension.width]);
    // 		yScale.domain([0, dimension.height]);
    // 		clickedItem = NULL;
    // 	} else {
    // 		xScale.domain([d.x0, d.x1]);
    // 		yScale.domain([d.y0, d.y1]);
    // 		clickedItem = getLevel(d);
    // 	}
    // 	chart.defaultMouseOutFunc();
    // 	$ZC.labelRenderer.drawLabelsFromQueue(rp.labelQueue, rp.labelArray, chartObj);
    // 	path
    // 	.transition()
    // 	.ease(d3_ease_linear)
    // 	.attr("d", function(d) {
    // 		var tileBound = treemapUtils.getTileBound(chartObj, d);
    // 		return roundedRect(tileBound.x, tileBound.y, tileBound.width, tileBound.height, borderRadius)
    // 	});
    // }
    chart.getProcessedNode = function(d) {
        return layoutObj.nodes[getLevel(d)];
    };

    chart.getHighlighter = function(chartObj, index) {
        return (chart.highlighter = new TreemapHighlighter(chartObj, chart));
    };

    chart.destroy = function destroy() {
        path = treemapGroup = null;
    }

    return chart;
};

export var _treemapChart = svg_plot_renderer.treemap;
