// @zohocharts/zc/src/treemap/treemapOptions.js
import { defaultTheme as options } from '../core/defaulttheme';

options.treemap = function() {
    return {
        outerPadding: 5,
        strokeColor: 'rgba(255,255,255,0.7)',
        animation: {
            type: 'fadeOneByOne'
        },
        axis: {
            "enabled": false
        },
        datalabels: {
            showAs: 'y',
            vAlign: 'bottom',
            hAlign: 'left'
        },
        gradients: {
            //linear | radial | none
            options: {
                linear: {
                    y2: 100,
                    x2: 100,
                    colorGamma: [0.5, 0]
                },
                radial: {
                    radius: 100
                }
            }
        },
        parent: {
            strokeOpacity: 0,
            datalabels: {
                position: 'inset',
                vAlign: 'center',
                hAlign: 'center',
                showAs: 'y',
                show: false
            },
            events: {}
        }
    };
};

export var _treemapOptions = options.treemap;
