// @zohocharts/zc/src/axis/cartesian/axis-marks.js
import { applyFont } from '../../utils/zcsvgutils';
import { $ZCG } from '../../core/zcglobal';
//TODO: Move to registry and add in js_list
import { _cartesianDomain as domain } from './parser/marks/cartesian-domain-marks';
import { _cartesianTicks as ticks } from './parser/marks/cartesian-ticks-marks';
import { _cartesianSubMinorTicks as subminorticks } from './parser/marks/cartesian-subminorticks';
import { _cartesianSubMajorTicks as submajorticks } from './parser/marks/cartesian-submajorticks';
import { _cartesianAxisLabel as axislabel } from './parser/marks/cartesian-axislabel-marks';
import { ParserRegistry } from '../../core/parser-registry';

export function _defaultAxisMark() {
    let _ = this,
        styleObj = applyFont(null, [_.config.ticklabel, _.chart.systemConf.chart, _.chart.systemConf.canvas, $ZCG]);

    this.helpers = {};
    this.helpers.styles = styleObj;

    return function(positionH, axisTitleH, component, transition_, container) {
        // TODO: Later move to class structure initialization
        _.helpers.position = positionH;
        _.helpers.title = axisTitleH;
        _.component = component;
        _.helpers.transition = transition_;
        _.container = container;

        const domainMarks = domain.call(_);
        const ticksMarks = ticks.call(_);
        const subMinorTicksMarks = subminorticks.call(_);
        const subMajorTicksMarks = submajorticks.call(_);
        const axisLabelMarks = axislabel.call(_);

        let _axisMark = {
            renderOnDemand: true,
            datatype: _.datatype,
            column: _.column,
            scale: {
                accessor: () => {
                    // return viewParser.encodes[columnIdx].scale;
                }
            },
            order: _._order,
            config: _.config,
            orient: _.orient,
            mark: {
                mark: 'g',
                selector: '.axisgroups',
                data: [1],
                children: [
                    domainMarks,
                    ticksMarks,
                    subMinorTicksMarks,
                    subMajorTicksMarks,
                    axisLabelMarks
                ]
            }
        };
        return _axisMark;
    };
}
ParserRegistry.axis('cartesian.default', _defaultAxisMark);
