// @zohocharts/zc/src/axis/cartesian/axis-prerender-marks.js
import { applyFont } from '../../utils/zcsvgutils';
import { $ZCG } from '../../core/zcglobal';
import { AxisRegistry } from '../axis-registry';

export function _defaultAxisPreRenderMark() {
    let _ = this,
        styleObj = applyFont(null, [_.config.ticklabel, _.chart.systemConf.chart, _.chart.systemConf.canvas, $ZCG]),
        axisTitleH;
    return function(positionH) {
        let _x_tl = positionH.ticklabel.attrs_key.x, //tl -ticklabel, tm- tickmark
            _y_tl = positionH.ticklabel.attrs_key.y,
            _x_tm = positionH.tickmark.attrs_key.x;
        let _axisMark = {
            renderOnDemand: true,
            datatype: _.datatype,
            column: _.column,
            scale: {
                accessor: () => {
                    // return viewParser.encodes[columnIdx].scale;
                }
            },
            order: _._order,
            config: _.config,
            orient: _.orient,
            mark: {
                mark: 'g',
                selector: '.axisgroups',
                data: [1],
                properties: {
                    attrs: {
                        transform: positionH.axisG.transform
                    }
                },
                children: [
                    {
                        mark: 'g',
                        selector: '.zc-tick',
                        data: [_.axisTicks.tickInfo.ticks_formatted[0]],
                        properties: {
                            attrs: {
                                transform: function(d) {
                                    let useIndices = _.axisTicks.tickInfo.scale.discreteWidth || _.axisTicks.useTickAsIndices,
                                        d1 = useIndices ? _.axisTicks.tickInfo.scale.domain()[0] : d;
                                    return positionH.tickG.transform(d1);
                                }
                            }
                        },
                        children: [
                            {
                                name: 'ticklabel',
                                selector: '.ticklabel',
                                mark: 'text',
                                datum: () => {},
                                data: [_.axisTicks.tickInfo.ticks_formatted[0]],
                                properties: {
                                    text: function(d, i) {
                                        return d;
                                    },
                                    attrs: {
                                        [_x_tl]: positionH.ticklabel[_x_tl],
                                        [_y_tl]: positionH.ticklabel[_y_tl],
                                        dy: positionH.ticklabel.dy,
                                        transform: positionH.ticklabel.transform
                                    },
                                    styles: {
                                        'text-anchor': positionH.ticklabel.textAnchor,
                                        ...styleObj
                                    }
                                }
                            },
                            {
                                name: 'tickmark',
                                selector: '.tickmark',
                                mark: 'line',
                                data: [_.axisTicks.tickInfo.ticks_formatted[0]],
                                properties: {
                                    attrs: {
                                        [`${_x_tm}1`]: positionH.tickmark[`${_x_tm}1`],
                                        [`${_x_tm}2`]: positionH.tickmark[`${_x_tm}2`]
                                    }
                                }
                            }
                        ]
                    },
                    {
                        name: 'axisline',
                        selector: '.axisline',
                        mark: 'line',
                        data: [1],
                        properties: {
                            attrs: {
                                x1: positionH.axisline.x1,
                                y1: positionH.axisline.y1,
                                x2: positionH.axisline.x2, // phantomjs 1.9.8 issue
                                y2: positionH.axisline.y2 // // phantomjs 1.9.8 issue
                            }
                        }
                    }
                ]
            },
            axisTitle: {
                onReady: function() {
                    axisTitleH = AxisRegistry.get('cartesian.axisTitle').call(_);
                    return {
                        mark: 'g',
                        selector: '.axislabelgroup',
                        data: [1],
                        children: [
                            {
                                name: 'axislabel',
                                selector: '.axislabel',
                                mark: 'text',
                                data: [axisTitleH.text],
                                properties: {
                                    text: function(d, i) {
                                        return _.axisTitle.hideTitle ? '' : d;
                                    },
                                    attrs: {
                                        x: axisTitleH.x,
                                        y: axisTitleH.y,
                                        dy: axisTitleH.dy,
                                        transform: axisTitleH.transform
                                    },
                                    styles: {
                                        'text-anchor': axisTitleH.textAnchor,
                                        ...axisTitleH.styleObj
                                    }
                                }
                            }
                        ]
                    };
                }
            }
        };
        return _axisMark;
    };
}

AxisRegistry.add('cartesian.axisPreRenderMarks', _defaultAxisPreRenderMark);
