// @zohocharts/zc/src/axis/cartesian/axis-title-render.js
import { getAxisAreaDim } from '../hierarchy-axis-utils';
import { getAxisObject } from '../axisutils';
import { appendEle } from '../../utils/zcsvgutils';
import { axis } from '../../core/main-zc-init';

export function initializeAxisTitle(axisTitleParser) {
    let titleH = (this.titleH = axisTitleParser.call(this, this.axisTicks.tickH));
    return {
        hideTitle: titleH.hideAxislabel(),
        titleConfig: titleH.axislabelData(),
        titleMargin: titleH.margin(),
        branchTitleConfig: titleH.branchLabelData()
    };
}
export function renderAxisTitle(placeHolder) {
    var isHorizontal = this.type === 'x' && !this.rotated,
        isVertical = !isHorizontal,
        axisLabelProperties = {
            systemConf: systemConf,
            bound: axisgroup,
            horizontal: isHorizontal,
            vertical: isVertical,
            axisOrder: this.order,
            plotRect: this.plotRect,
            hide: this.hideAxisLabel,
            tickObj: this.tickPrecalculation,
            type: this.type,
            chartObj: chart
        },
        axislabelProp = { textAnchor: 'middle' };

    var axisSize = 0;
    var axisdata_merged = getBranchLabelData(helpers),
        axes = getAxisObject(chart, this.type, this.order),
        selector = this.type == 'x' ? `${this.type}-axis-${this.order}` : `${this.type}-axis-${this.order}`,
        axisG = placeHolder,
        axisWrap = appendEle(axisG, 'g', [this.config], '.', 'class', selector);
    if (axes.axisLabel && axes.axisLabel.order == this.order) {
        let axisAreaDim = getAxisAreaDim(this.type, this.order, this.orient, chart, this.plotRect);
        axes.axisLabel.show = !this.hideAxisLabel;
        axes.axisLabel.transformLabel(axisgroup, axisAreaDim, this.orient);
    } else {
        axes.axisLabel = new axis.AxisLabel(axisdata_merged, axisLabelProperties, this.orient, axisWrap);
    }
}
