// @zohocharts/zc/src/axis/cartesian/cartesian-axis-prerender.js
import { AxisTicks } from '../ticks/axis-ticks';
import { precalculateAxisGroup } from '../ticks/axis-group-precalculation';
import { precalculateTickGroup } from '../ticks/longest-tick';
import { getPositionForTicks } from './ticks-position-finder';
import { initializeAxisTitle } from './axis-title-render';
import { AxisRegistry } from '../axis-registry';
import { invokeFunction } from '../../utils/zcutils';

export class CartesianAxisPrerender {
    constructor(args, helpers) {
        let { type, order, config, plotRect, viewParser, maxSize, chart } = args;
        this.type = type;
        this.order = order;
        this.config = config;
        this.plotRect = plotRect;
        this.plotRect1 = plotRect;
        this.maxSize = maxSize;
        this.chart = chart;

        this.initialize(args, helpers);
    }
    initialize(args, helpers) {
        let _h = helpers.axisH;
        this._order = _h.order(this.type, this.order);
        this.orient = _h.orient(this.type, this.order);
        this.datatype = _h.datatype(this.type, this.order);
        this.column = _h.column(this.type, this.order);
        this.axisInfo = _h.axisInfo(this.type, this.order);
        this.tickParser = helpers.tickH;
        this.titleParser = helpers.titleH;

        this.axisView = this.axisInfo.axisView;
    }

    getDefaultMarkParser() {
        let markParser = 'cartesian.axisPreRenderMarks';
        if (this.axisView && this.axisView.conf && this.axisView.conf.preMarkParser) {
            markParser = this.axisView.spec.preMarkParser;
        }
        return {
            placeHolder: this.chart.svg.append('g').attr('class', 'dummy'),
            markParser: AxisRegistry.get(markParser).call(this)
        };
    }
    title() {
        let hideTitle = this.titleH.hideTitle(),
            titleConfig = this.titleH.titleConfig();
        renderAxisTitle.call(this);
    }
    getTicks(newPlotRect) {
        let args = {
            chart: this.chart,
            type: this.type,
            _order: this._order,
            config: this.config,
            plotRect: this.plotRect,
            plotRect1: this.plotRect1,
            order: this.order,
            orient: this.orient,
            datatype: this.datatype,
            column: this.column,
            maxSize: this.maxSize,
            axisInfo: this.axisInfo
        };
        this.axisTicks = new AxisTicks(args, this.tickParser);
        this.axisTicks.collect();
    }
    preRender(newPlotRect) {
        this.plotRect = newPlotRect || this.plotRect;
        this.preRenderTickG();
        this.preRenderAxisG();
    }
    preRenderTickG(newPlotRect) {
        this.plotRect = newPlotRect || this.plotRect;
        this.existingTickInfo = this.axisTicks && this.axisTicks.tickInfo ? this.axisTicks : null;
        this.getTicks();
        let positionH = getPositionForTicks.call(this);
        let markExtender = this.axisView && this.axisView.conf.preMarkExtender;
        let markInfo = this.getDefaultMarkParser(),
            marks = markInfo.markParser(positionH),
            mark = marks.mark;
        this.marks = marks;
        if (markExtender) {
            mark = invokeFunction(markExtender, mark, this);
        }
        var tempDom = (this.placeHolder = markInfo.placeHolder);
        this.axisTicks.precalculate = precalculateTickGroup.call(this, mark, tempDom);
        this.placeHolder.remove();
        this.placeHolder = null;
    }
    preRenderAxisG() {
        this.axisTitle = initializeAxisTitle.call(this, this.titleParser);
        // this.title();
        this.axisTitle.precalculate = precalculateAxisGroup.call(this);
    }
    destroy() {
        this.marks = null;
    }
}

AxisRegistry.add('cartesian.prerender', CartesianAxisPrerender);
