// @zohocharts/zc/src/axis/cartesian/ticks-position-finder.js
import { LOG } from '../../core/scale-type-constant';
import { defined, isArray, pick, validateNullVal } from '../../utils/zcutils';
import { checkTickInlineAlignWithInPlot, getMargin, getTickPadding, isTickInnerOrient } from '../axisutils';
import { mathAbs } from '../../core/zc-constants';
import { getTickRotateInfo } from '../tickRotate-helpers';
import { AxisRegistry } from '../axis-registry';

/*
render -> true -> render
render -> false -> prerender
 */
export function getPositionForTicks(render, scale) {
    let _ = this,
        { existingTickInfo } = _,
        oldTickObj = existingTickInfo && existingTickInfo.precalculate,
        { tickInfo } = _.axisTicks,
        attrs_key = {
            x: _.orient === 'left' || _.orient === 'right' ? 'x' : 'y',
            w: _.orient === 'left' || _.orient === 'right' ? 'width' : 'height',
            y: _.orient === 'left' || _.orient === 'right' ? 'y' : 'x'
        },
        attrs_value = {
            k: _.orient === 'top' || _.orient === 'left' ? -1 : 1
        };

    /*
    Margin, padding  details
     */
    let ticklabelMargin = getMargin(_.config.ticklabel),
        axislabelMargin = getMargin(_.config.label),
        tickPadding = getTickPadding(_.orient, ticklabelMargin, isTickInnerOrient(_.config, _.type, _.chart)),
        tickSizeInner = validateNullVal(_.config.tickmark.size, 6),
        supportSubTicks = _.column.scaleType === LOG,
        /*
        Sub minor ticks details, currently added for log scale
         */
        minorGrid = tickInfo.minorTickValues,
        hasMinorGridPositions = supportSubTicks || (isArray(minorGrid) && minorGrid.length > 0),
        subTickMarkData = _.config.subtickmark || {},
        isSubTickMarkShow = (subTickMarkData.count > 0 || defined(subTickMarkData.interval)) && hasMinorGridPositions,
        subTickSize = isSubTickMarkShow ? validateNullVal(_.config.subtickmark.size, 6) : 0,
        tickmarkSize = Math.max(tickSizeInner, subTickSize, 0),
        /*
        ticklabel default spacing with tickmark size and padding
         */
        spacing_default = tickmarkSize + tickPadding,
        /*
         ticklabel as inline mode
         */
        isInline = isTickInnerOrient(_.config, _.type, _.chart),
        inlineToPlot = isInline ? checkTickInlineAlignWithInPlot(_.type, _.orient, _.chart.systemConf) : false,
        inlineWspace = isInline && !inlineToPlot,
        inline_reRender = oldTickObj && inlineWspace,
        /*
        controllers
         */
        l = inlineToPlot ? -attrs_value.k : attrs_value.k,
        m = isInline ? -attrs_value.k : attrs_value.k,
        /*
        tick label positioning:
        For normal ticks, it will take from tickmark size and padding
        For inlineWspace charts, on init, consider only tickpadding, because tickmark size should not be included in precalculation
        For final rendering cnsider tickmarksize, padding
         */
        spacing = inline_reRender ? oldTickObj.longTickSize[attrs_key.w] : inlineWspace ? tickPadding : spacing_default,
        /*
        tickmark positioning
         */
        tickmarkStart = inline_reRender ? oldTickObj.axisgroup[attrs_key.w] : 0,
        /*
        rotation details
         */
        tickRotation = pick(tickInfo.rotation, 0),
        rotation = tickRotation < 0 ? 360 - mathAbs(tickRotation) : tickRotation;
    rotation = tickRotation > 360 ? mathAbs(rotation) - 360 : rotation;

    let tickObj = oldTickObj ? { ...existingTickInfo.tickInfo, ...existingTickInfo.precalculate } : null;
    let rotationInfo = getTickRotateInfo(_.orient, rotation || 0, null, null, tickmarkSize, tickPadding, tickObj); //handling subticksize also;

    let positionH = {
        attrs_key,
        attrs_value,
        m,
        l,
        spacing,
        tickmarkStart,
        rotation,
        rotationInfo,
        tickSizeInner,
        subTickSize
    };

    return {
        axisG: AxisRegistry.get('cartesian.axisGPosition').call(this),
        tickG: AxisRegistry.get('cartesian.tickGPosition').call(this, scale),
        ticklabel: AxisRegistry.get('cartesian.ticklabelPosition').call(this, positionH),
        tickmark: AxisRegistry.get('cartesian.tickmarkPosition').call(this, positionH, positionH.tickSizeInner, render),
        subtickmark: AxisRegistry.get('cartesian.tickmarkPosition').call(this, positionH, positionH.subTickSize, render),
        gridline: _.axisInfo && _.axisInfo.ticks ? _.axisInfo.ticks.majorTicks : null,
        axisline: AxisRegistry.get('cartesian.axisLinePosition').call(this),
        axislinePath: AxisRegistry.get('cartesian.axisLinePath').call(this, positionH)
    };
}

AxisRegistry.add('cartesian.positionFinder', getPositionForTicks);
