// @zohocharts/zc/src/legend/size-legend/size-legend-options.js
import { TOP_ALIGN, RIGHT_ALIGN, VERTICAL_ALIGN } from '../../core/align-constants';
import { TRANSPARENT } from '../../core/zc-constants';
import { Registry } from '../../core/zc-registry';
import { defaultTheme as options } from '../../core/defaulttheme';
import { brushTooltipOptions } from '../legend-brush-options';

options.sizeLegend = function(userProp) {
    let nestedOptions = Registry.getComponent('sizeNestedOptions')(userProp),
        listOptions = Registry.getComponent('sizeListOptions')(userProp),
        sliderOptions = Registry.getComponent('sizeSliderOptions')(userProp);

    return {
        enabled: true,
        hAlign: RIGHT_ALIGN, //left, right, center
        vAlign: TOP_ALIGN, //top, center, bottom
        layout: VERTICAL_ALIGN,
        itemMargin: 3,
        marginTop: 10,
        marginRight: 10,
        marginBottom: 10,
        marginLeft: 10,
        title: {
            show: false,
            text: 'Legend',
            fontSize: userProp.fontL1Size.l4
        },
        background: {
            color: TRANSPARENT,
            alpha: 0
        },
        border: {
            show: true,
            size: 1,
            style: 'solid',
            radius: 0,
            color: TRANSPARENT
        },
        shadow: {
            show: false,
            color: 'rgba(0,0,0,0.3)',
            x: 2,
            y: 2,
            blur: 3,
            inset: false
        },
        highlightEffect: {
            selectedSeries: null,
            unselectedSeries: 'none'
        },
        list: listOptions,
        nested: nestedOptions,
        slider: sliderOptions,
        hoverOptions: {
            slider: {
                tooltip: {
                    fontSize: userProp.fontL1Size.l5
                }
            }
        },
        filter: {
            enabled: true,
            slider: {
                track: {},
                tooltip: {
                    ...brushTooltipOptions(),
                    fontSize: userProp.fontL1Size.l5
                }
            }
        }
    };
};

// {
//     enabled: true,
//     title: {
//         show: false,
//         text: 'Legend',
//         fontSize: userProp.fontL1Size.l4
//     },
//     box: {
//         shape: 'circle',
//         borderRadius: null,
//         hAlign: 'center',
//         vAlign: null,
//         fillColor: '#EBEBEB',
//         fillOpacity: null,
//         strokeColor: '#ABABAB',
//         strokeWidth: null,
//         strokeOpacity: null,
//         text: null,
//         fontColor: null,
//         fontFamily: null,
//         fontSize: null,
//         fontWeight: null,
//         fontStyle: null
//     },
//     label: {
//         fontColor: null,
//         fontFamily: null,
//         fontSize: userProp.fontL1Size.l5,
//         fontWeight: null,
//         fontStyle: null,
//         showAs: null,
//         padding: 5,
//         keepAligned: true,
//         overflow: null, //check overflow wrap
//         tooltip: null
//     },
//     nested: {
//         shape: 'circle',
//         hAlign: 'center',
//         vAlign: 'bottom',
//         fillColor: 'transparent',
//         strokeColor: '#707070'
//     },
//     axis: {
//         show: true,
//         ticklabel: {
//             margin: null,
//             fontFamily: null,
//             fontSize: 9 //userProp.fontL1Size.l5
//         },
//         tickmark: {
//             show: false,
//             color: '#707070',
//             size: 3,
//             strokeWidth: null
//         }
//     },
//     background: {
//         color: 'transparent',
//         alpha: 0
//     },
//     border: {
//         show: true,
//         size: 1,
//         style: 'solid',
//         radius: 0,
//         color: 'transparent'
//     },
//     shadow: {
//         show: false,
//         color: 'rgba(0,0,0,0.3)',
//         x: 2,
//         y: 2,
//         blur: 3,
//         inset: false
//     },
//     hAlign: 'right', //left, right, center
//     vAlign: 'top', //top, center, bottom.
//     itemMargin: 3,
//     marginTop: 10,
//     marginRight: 10,
//     marginBottom: 10,
//     marginLeft: 100
// };

export let _sizeLegendOptions = options.sizeLegend;
