// @zohocharts/zc/src/legend/shape-legend/shape-legend-options.js
//
import { TOP_ALIGN, RIGHT_ALIGN, VERTICAL_ALIGN } from '../../core/align-constants';
import { TRANSPARENT } from '../../core/zc-constants';
import { Registry } from '../../core/zc-registry';
import { defaultTheme as options } from '../../core/defaulttheme';

options.shapeLegend = function(userProp) {
    let listOptions = Registry.getComponent('shapeListOptions')(userProp);

    return {
        enabled: true,
        hAlign: RIGHT_ALIGN, //left, right, center
        vAlign: TOP_ALIGN, //top, center, bottom
        layout: VERTICAL_ALIGN,
        itemMargin: 3,
        marginTop: 10,
        marginRight: 10,
        marginBottom: 10,
        marginLeft: 10,
        title: {
            show: false,
            text: 'Legend',
            fontSize: userProp.fontL1Size.l4
        },
        background: {
            color: TRANSPARENT,
            alpha: 0
        },
        border: {
            show: true,
            size: 1,
            style: 'solid',
            radius: 0,
            color: TRANSPARENT
        },
        shadow: {
            show: false,
            color: 'rgba(0,0,0,0.3)',
            x: 2,
            y: 2,
            blur: 3,
            inset: false
        },
        highlightEffect: {
            selectedSeries: null,
            unselectedSeries: 'none'
        },
        list: listOptions,
        hoverOptions: {},
        filter: {
            enabled: true
        }
    };
};

export let _shapeLegendOptions = options.shapeLegend;
