// @zohocharts/zc/src/behaviours/data-selection/chart/tools/selection-to-domain.js
import { getOrdinalRange } from '../../../../axis/scale';
import { ORDINAL_DATATYPE } from '../../../../core/datatype-constants';
import { scale } from '../../../../core/main-zc-init';
import { mathMax, mathMin } from '../../../../core/zc-constants';
import { defined, getPropVal, isArray, isUndefined } from '../../../../utils/zcutils';

/**
 *
 * @param selection
 * @param chartScale
 * @param type
 * @param datatype
 * @param rotated
 * @param reversed
 * @return {*[]}
 * @constructor
 */

export function SelectionToDomain(selection, chartScale, type = 'x', datatype, rotated = false, reversed = false, chartObj, isBrush) {
    var eventHandler = chartObj.eventHandler,
        selectionDomain = eventHandler.selectionDomain,
        selectionDomainByType = selectionDomain && selectionDomain[type];
    if (!isArray(selection)) {
        return chartScale.originalDomain;
    }

    if(selectionDomainByType && selectionDomainByType.length && !(eventHandler.activeBrushComponent === "plot")){
        return selectionDomainByType;   
    }
    let domain,
        selectionInfo = {
            selection: selection,
            quadTreePosition: []
        },
        [inverted_0, inverted_1] = selection.map((x0, i) => {
            let a = { x: 0, y: 0 };
            a[fixBrushDirection(type, rotated)] = x0;
            return scale.invert(datatype, a, chartScale, rotated, type, reversed, null, selectionInfo, chartObj, isBrush);
        });
    if (datatype === ORDINAL_DATATYPE) {
        var [_s, _e] = selectionInfo.quadTreePosition,
            ordinalRange = getOrdinalRange(chartScale),
            rangeBand = chartScale.bandwidth(),
            _curr_pos = ordinalRange[inverted_1] + rangeBand / 2,
            _prev_pos = ordinalRange[inverted_1 - 1] + rangeBand / 2; //getting the middle point of the bar/marker
        if (!selectionInfo.quadTreePosition.length) {
            if (isEmptySelection(selection, ordinalRange, inverted_0, rangeBand)) {
                [inverted_0, inverted_1] = [-1, 0];
            }
        } else if (isUndefined(_s) && isUndefined(_e)) {
            if (!(_curr_pos >= selection[0] && _curr_pos <= selection[1])) {
                if(!(_prev_pos >= selection[0] && _prev_pos <= selection[1])){
                    [inverted_0, inverted_1] = [-1, 0];
                }
            }
        }
    }

    if (inverted_0 > inverted_1) {
        [inverted_0, inverted_1] = [inverted_1, inverted_0];
    }

    if (datatype === ORDINAL_DATATYPE) {
        const scrollInstance = getPropVal(chartObj, 'behaviouralEvents.eventManager.behaviours.CHART_SCROLL.instance'),
            isOverviewEnabled = getPropVal(scrollInstance, 'overviewEnabled');
        if (!isBrush && !isOverviewEnabled && !chartObj.racing) {
            const visiblePointCount = getPropVal(scrollInstance, 'visiblePointCount');
            inverted_1 += defined(visiblePointCount) ? visiblePointCount - (inverted_1 - inverted_0 + 1) : 0;
        }
        domain = chartScale.domain().slice(inverted_0, inverted_1 + 1, 0);  // changed +1 to +2 //ZC4389
    } else {
        domain = [inverted_0, inverted_1]; //constrainDomain([inverted_0, inverted_1], chartScale.originalDomain, chartScale.outerPadding);
    }
    
    if(eventHandler.oldBrushComponent === "plot" && eventHandler.activeBrushComponent === "scroll"){
        eventHandler.oldBrushComponent = null;
        eventHandler.activeBrushComponent = null;
    }//old and active brush components maintained inorder to avoid domain calculation multiple times when brushing is made.
    if(eventHandler.activeBrushComponent === "plot"){
        eventHandler.selectionDomain = selectionDomain ? selectionDomain : {};
        eventHandler.selectionDomain[type] = domain;
    }
    return domain;
}

/**
 *
 * @param axisType
 * @param rotated
 * @return {*}
 */
export function fixBrushDirection(axisType, rotated) {
    let dir_opposite_map = { x: 'y', y: 'x' };
    if (rotated) {
        return dir_opposite_map[axisType] || axisType;
    }
    return axisType;
}

export function constrainDomain(domain, ordinalDomain, outerPadding = 0) {
    let isDate = domain[0] instanceof Date;

    if (isDate) {
        domain = domain.map(t => t.getTime());
        ordinalDomain = ordinalDomain.map(t => t.getTime());
    }

    let [start, end] = domain,
        [startLimit, endLimit] = ordinalDomain,
        startDiff = (end - startLimit) * outerPadding,
        newStart = startLimit - startDiff,
        newEnd = endLimit + (endLimit - start) * outerPadding,
        // constrained
        s0 = mathMax(start, newStart),
        s1 = mathMin(end, newEnd);

    // fix if beyond start limit
    if (end < startLimit) {
        s0 = startLimit - (s0 - startLimit) * 0.001;
        s1 = startLimit + (startLimit - s1) * 0.001;
    }

    // fix if beyond end limit
    if (start > endLimit) {
        s0 = endLimit + (endLimit - s0) * 0.001;
        s1 = endLimit - (s1 - endLimit) * 0.001;
    }

    if (isDate) {
        [s0, s1] = [s0, s1].map(t => new Date(t));
    }

    return [s0, s1];
}

/**
 *
 * @param range
 * @param originalRange
 * @return {number[]}
 */
export function constrainRange(range, originalRange) {
    let [start, end] = originalRange;
    if (originalRange[0] < originalRange[1]) {
        range = [mathMin(range[0], start), mathMax(range[1], end)];
    } else {
        range = [mathMax(range[0], start), mathMin(range[1], end)];
    }
    return range;
}

function isEmptySelection(selection, ordinalRange, actPos, bandWidth) {
    var _s = selection[0], //start
        _e = selection[1], //end
        curr_Pos = ordinalRange[actPos],
        next_Pos = ordinalRange[actPos + 1],
        prev_Pos = ordinalRange[actPos - 1],
        curr_Pos_end = curr_Pos + bandWidth,
        next_Pos_end = next_Pos + bandWidth,
        prev_Pos_end = prev_Pos + bandWidth;
    function isWithin(_d, start, end) {
        if (_d >= start && _d <= end) {
            return true;
        }
        return;
    }
    if (
        isWithin(curr_Pos, _s, _e) ||
        isWithin(curr_Pos_end, _s, _e) ||
        isWithin(next_Pos, _s, _e) ||
        isWithin(next_Pos_end, _s, _e) ||
        isWithin(prev_Pos, _s, _e) ||
        isWithin(prev_Pos_end, _s, _e) ||
        isWithin(_s, curr_Pos, curr_Pos_end) ||
        isWithin(_e, curr_Pos, curr_Pos_end)
    ) {
        return false;
    }
    return true;
}
