// @zohocharts/zc/src/highlighter/arc-highlighter.js
import { d3_namespaces, d3_polygonCentroid } from "../core/d3-constants";
import { HIGHLIGHT_CONTEXT } from "../events/visualEffects/highlight/highlight-context";
import { patternFillEffect } from "../events/visualEffects/highlight/pattern-fill";
import { markSelection } from "../mark/extender/markSelection";
import { arcMark } from "../mark/helper/custom-mark/arc-mark";
import { translateString } from "../utils/svg-transforms";
import { appendEle, pathToPolygon } from "../utils/zcsvgutils";
import { defined, getPropVal } from "../utils/zcutils";

export default class ArcHighlighter {
    #rp; #sampleStep = 10;
    #tempPath = document.createElementNS(d3_namespaces.svg, 'path');
    constructor(chart, renderer) {
        this.chart = chart;
        this.renderer = renderer;
        this.#rp = renderer.commonRendererProp().params;
    }

    _defaultH() {
        let defaultH = [
            {
                shape: 'poly',
                path: this.path,
                data: this.data,
                context: this.renderingContext,
                translate: this.translateContext,
                styles: point => {
                    return patternFillEffect.call(this, point, 'inherit');
                }
            }
        ];
        return defaultH;
    }

    data(point) {
        const self = this;
        const { chart } = this;

        point._polydata = this.poly(point);

        return function _data() {
            const cache = chart.cache.data[point._key] || (chart.cache.data[point._key] = {});
            if (!defined(cache._x)) {
                const polygonPoints = self.getPolygon(point);
                const [x, y] = d3_polygonCentroid(polygonPoints);
                cache._x = point._x = x;
                cache._y = point._y = y;
                cache._defined = point._defined = true;
                cache._point = point;
                cache._poly = point._poly = { points: polygonPoints };
            }
            cache._visible = point._visible = self.isVisible(point);
            return point;
        }

    }

    poly(point) {
        const self = this,
            chart = this.chart;

        return function _poly() {
            const cache = chart.cache.data[point._key] || (chart.cache.data[point._key] = {});
            if (!cache._poly) {
                cache._poly = point._poly = { points: self.getPolygon(point) };
            }
            return point._poly;
        }
    }
    getPolygon(point, zoom) {
        if (point.disabled) { return [[null, null]] };
        const config = this.getDefaultConfig();
        const rp = this.renderer.commonRendererProp().params,
            sampleSize = getPropVal(rp, 'commonPlotOptions.events.sample') || 50;

        let path = this.getPath(point, config),
        // $path = this.getTempPath(point, config),
            polygon = pathToPolygon(path, sampleSize);

        return polygon ?? [[null, null]];
    }
    getDefaultConfig() {
        const parser = this.renderer.parser(),
            marks = parser.marks(),
            selection = new markSelection(marks[0].mark).select('mark', '_arc').selection,
            config = selection.anime.update;

        return config;
    }

    getPath(point, config) {
        const arc = arcMark(config, {}),
            path = arc.config.d(point, point.itemIndex);
        return path;
    }
    getTempPath(point, config) {
        const path = this.getPath(point, config);
        this.#tempPath.setAttribute('d', path);
        return this.#tempPath
    }

    isVisible(point, from) {
        return !point.disabled;
    }

    sampleStep(step) {
        return arguments.length ? (this.#sampleStep = step, this) : this.#sampleStep;
    }

    // default highlighter dependencies
    renderingContext() {
        return HIGHLIGHT_CONTEXT.get('bar').call(this, this.chart);
    }
    translateContext() {
        const { left, top } = this.chart.plotarea;
        return [left, top];
    }
    path(point, model) {
        const self = this;
        point = point._data();

        const config = this.getDefaultConfig(),
            plotarea = this.chart.plotarea,
            plotarea_approx = this.chart.plotarea_approx,
            left = plotarea_approx.x + plotarea.left,
            top = plotarea_approx.y + plotarea.top,
            path = this.getPath(point, config);

        const context = model.context.call(this),
            pathContext = appendEle(context, 'path', [1], '.', 'class', `highlight${point.paneid}${point.seriesIndex}${point.subSeriesIndex}${point.itemIndex}`);

        context.style('display', '');
        pathContext.styles({
            display: '',
        }).attrs({
            d: path,
            transform: translateString([left, top]),
        })

        return pathContext;
    }
    options(point) {
        // const { chart, renderer } = this;
        return {
            color: 'transparent'
        }
    }
};
