// @zohocharts/zc/src/axis-parser/tick-parser-helpers/tick-rotation-helpers.js
import { defined } from '../../utils/zcutils';
import { hasBranchAxis, isBranchAxis, isCategoryHierarchical } from '../../axis/hierarchy-axis-utils';
import { mathAbs } from '../../core/zc-constants';

export function _tickRotation(obj) {
    var alignmentMode = obj.alignMode,
        isRotateMode = obj.alignMode && obj.alignMode.indexOf('rotate') != -1,
        isOldRotateMode = defined(obj.rotation),
        rotation = 0,
        hasPriorityOrder,
        autoRotation = 45,
        hasDefinedAutoRotation;
    if (isRotateMode) {
        var rotationExtract = /\(([^)]+)\)/.exec(alignmentMode);
        rotation = rotationExtract ? parseInt(rotationExtract[1]) : 'auto';
    } else if (isOldRotateMode && !defined(obj.alignMode)) {
        rotation = obj.rotation == 'auto' ? obj.rotation : parseInt(obj.rotation);
        (alignmentMode = 'rotate'), (isRotateMode = true);
    } else if (Array.isArray(alignmentMode) && alignmentMode.length) {
        hasPriorityOrder = true;
        alignmentMode.some(d => {
            if (d.indexOf('rotate') != -1) {
                alignmentMode = 'rotate';
                isRotateMode = true;
                var rotationExtract = /\(([^)]+)\)/.exec(d);
                rotation = 'auto';
                hasDefinedAutoRotation = rotationExtract;
                autoRotation = rotationExtract ? parseInt(rotationExtract[1]) : autoRotation;
                return true;
            }
        });
    } else if (!defined(obj.alignMode)) {
        (alignmentMode = 'rotate(0)'), (isRotateMode = true);
    }
    return {
        tickAlignmentMode: alignmentMode,
        tickRotateMode: isRotateMode,
        rotation: rotation,
        hasPriorityOrder: hasPriorityOrder,
        hasDefinedAutoRotation: hasDefinedAutoRotation,
        tickAutoRotation: autoRotation
    };
}

export function needTickAutoRotate(chartObj, type, order, rotationConfig) {
    var isBranchAxisType;
    if (chartObj.dataObject.categoryHierarchy && type == 'x') {
        isBranchAxisType = isBranchAxis(type, order, chartObj);
    }
    return rotationConfig == 'auto' && !isBranchAxisType;
}

export function _autoRotation(helpers) {
    let { type, order, chart, config } = helpers,
        tickConfig = config.ticklabel,
        rotationConfig = _tickRotation(tickConfig),
        tickRotateMode = rotationConfig.isRotateMode,
        tickRotation = rotationConfig.rotation,
        tickAutoRotation = rotationConfig.tickAutoRotation,
        hasDefinedAutoRotation = rotationConfig.hasDefinedAutoRotation,
        isAutoRotation = needTickAutoRotate(chart, type, order, tickRotation),
        categoryHierarchy = chart.dataObject.categoryHierarchy,
        isBranchAxisType = isCategoryHierarchical(chart) ? isBranchAxis(type, order, chart) : type == 'x' && order > 0,
        rotation = isBranchAxisType ? 0 : tickRotation,
        tickAlignmentMode = rotationConfig.alignmentMode,
        hasTickAlignmentPriority = rotationConfig.hasPriorityOrder;

    if (!isAutoRotation) {
        if (rotation < 0) {
            rotation = 360 - mathAbs(rotation);
        } else if (rotation >= 360) {
            rotation = mathAbs(rotation) % 360;
        }
    }
    if (tickAutoRotation) {
        if (
            (categoryHierarchy && type == 'x' && !isBranchAxisType && hasBranchAxis(type, helpers.chart)) ||
            (!categoryHierarchy && type == 'x' && helpers.chart.axes.x.count > 1)
        ) {
            tickAutoRotation = 90;
        } else {
            if (tickAutoRotation < 0) {
                tickAutoRotation = 360 - mathAbs(tickAutoRotation);
            } else if (tickAutoRotation > 360) {
                tickAutoRotation = mathAbs(tickAutoRotation) % 360;
            }
        }
    }
    rotationConfig.rotation = rotation;
    rotationConfig.tickAutoRotation = tickAutoRotation;
    return rotationConfig;
}
