// @zohocharts/zc/src/view/list/list-parser.js
//
import { TOP_ALIGN, LEFT_ALIGN, RIGHT_ALIGN, MIDDLE_ALIGN, HORIZONTAL_ALIGN, CENTER_ALIGN } from '../../core/align-constants';
import { NULL, NONE } from '../../core/zc-constants';
import { $ZCG } from '../../core/zcglobal';
import { mathMax } from '../../core/zc-constants';
import { d3_select } from '../../core/d3-constants';
import { Mark } from '../../mark/mark-utils';
import { applyFont } from '../../utils/zcsvgutils';
import { defined, pick, shorthand_px, mergeJSON, isEmptyObject, setPropVal, each } from '../../utils/zcutils';
// import { legend_helpers } from '../../legend/legend-helper';
import { legend_helpers } from '../../core/main-zc-init';
import { getColorBoxVerticalPosition } from '../../legend/legend-colorbox';
import { getTextHeight } from '../../label/font-helpers';
import { highlightLabel, highlightMarker, getListIndex } from '../../legend/events/visual-effects/size-list/highlight-item-helpers';
import { getDisabledOptions } from '../../legend/events/legend-event-helpers';

function ListParser(spec) {
    let _helper = helper(spec),
        { _class, show, data, ulSpec, liSpec, markerSpec, labelSpec, liSpec_onEnd, ulSpec_onEnd } = _helper,
        marker_P = MarkerParser(markerSpec),
        label_P = LabelParser(labelSpec);

    let ulMark = Mark('ul', `.${_class}legseries`);
    ulMark.properties = { styles: ulSpec };

    let liMark = Mark('li', `.${_class}series`, show ? data : []);
    liMark.properties = { 
        styles: liSpec,
        props: {
            __index__: (_, i) => i
        }
    };
    liMark.onEnd = liSpec_onEnd;
    liMark.children = [marker_P, label_P];

    ulMark.children = [liMark];
    ulMark.onEnd = ulSpec_onEnd;
    return ulMark;
}

function MarkerParser(spec) {
    let { show, bound, config, _class, onEnd, styles: _styles, order } = spec,
        staticConfig = { 'background-repeat': 'repeat', 'box-sizing': 'border-box', 'text-align': CENTER_ALIGN };

    let markerMark = Mark('_div', `.${_class}legendfilterbox`, show ? d => [d] : []);
    setPropVal(markerMark, 'properties.styles', { ...staticConfig, ..._styles });
    markerMark.onEnd = onEnd;
    markerMark.depends = true;
    markerMark.config = { ...config, width: bound.width, height: bound.height };
    markerMark.order = order;

    return markerMark;
}

function LabelParser(spec) {
    let { _class, onEnd, config, format = d => d, styles: _styles } = spec;

    let labelMark = Mark('span', `.${_class}legendlabel`, d => [d]);
    setPropVal(labelMark, 'properties.styles', _styles);
    labelMark.onEnd = onEnd;
    labelMark.depends = true;
    labelMark.config = { ...config, html: format };

    return labelMark;
}

function helper(spec) {
    let { data, config, align, format = d => d, globalFont = {}, class: _class = '', scales = {}, computedInfo } = spec, // Remove class
        globalFontSize = spec.label.fontSize,
        { list: listConfig, show = true } = config,
        { marker: markerConfig, label: labelConfig } = listConfig,
        boxSize = getBoxSize(markerConfig, labelConfig, align, scales.size, globalFontSize),
        place = _place(spec),
        _styles_marker = computedInfo && computedInfo.marker ? computedInfo.marker.config : markerConfig,
        _styles_label = computedInfo && computedInfo.label ? computedInfo.label.config : labelConfig;
        //computed styeles will not be there for SIZE LEGENDS

    return {
        show,
        data,
        _class,
        ulSpec: {...parseUL(config, align), ...(computedInfo && computedInfo.ul && computedInfo.ul.styles)},
        liSpec: {
            ...parseLI(config, align, place.li, globalFont),
            ...(computedInfo && computedInfo.li && computedInfo.li.styles)
        },
        markerSpec: {
            show: isBoxEnabled(markerConfig),
            config: {
                ...configByEncode(scales),
                ..._styles_marker
            },
            bound: boxSize,
            _class,
            ...place.marker,
            order: !markerConfig.position ? `.${_class}legendlabel` : null
        },
        labelSpec: {
            config: { ..._styles_label},
            format,
            globalFont,
            _class,
            ...place.label
        },
        liSpec_onEnd: computedInfo && computedInfo.li && computedInfo.li.onEnd,
        ulSpec_onEnd: computedInfo && computedInfo.ul && computedInfo.ul.onEnd
    };
}

function parseUL(align) {
    let isHorizontal = align.layout === HORIZONTAL_ALIGN;

    return {
        width: '',
        margin: 0,
        padding: 0,
        'white-space': isHorizontal ? 'nowrap' : NULL,
        [isHorizontal ? 'overflow-x' : 'overflow-y']: 'hidden',
        'list-style-type': NONE,
        '-webkit-overflow-scrolling': 'touch',
        'vertical-align': TOP_ALIGN,
        'font-size': '0px'
    };
}

function parseLI(config, align, place) {
    let labelConfig = config.list.label,
        itemWithIsNumber = typeof labelConfig.itemWidth === 'number';
    //text-overflow properties
    let overflowStyles;
    if (labelConfig.textOverflow === 'wrap') {
        overflowStyles = {
            'word-wrap': 'break-word',
            'white-space': 'normal',
            'vertical-align': TOP_ALIGN
        };
    } else {
        overflowStyles = {
            'text-overflow': 'ellipsis',
            'vertical-align': MIDDLE_ALIGN
        };
    }

    return {
        'white-space': 'nowrap',
        position: 'relative',
        overflow: 'hidden',
        'text-align': LEFT_ALIGN,
        padding: shorthand_px(pick(config.itemMargin, 3)),
        height: defined(config.itemHeight) ? `${config.itemHeight}px` : NULL,
        ...overflowStyles, //have to check this should be placed under li or span?
        'vertical-align': TOP_ALIGN,
        display: d => (legend_helpers.isHiddenItem(d) ? NONE : align.layout === HORIZONTAL_ALIGN ? 'inline-block' : 'block'),
        ...place.styles,
        'font-size': '0px',
        ...applyFont(NULL, [labelConfig], true), //check
        width: itemWithIsNumber ? `${labelConfig.itemWidth}px` : null//check
    };
}

function _place(spec) {
    let { data, config, align, ticks, globalFont, scales = {}, computedInfo, unitTestObj } = spec,
        globalFontSize = spec.label.fontSize,
        { list: listConfig, ranges: userRanges } = config,
        { marker: markerConfig, label: labelConfig } = listConfig,
        disabledOptions = getDisabledOptions(config),
        _defaultBoxSize = legend_helpers.getFilterBoxSize(markerConfig, globalFontSize),
        bound = spec.bound || { width: _defaultBoxSize[0], height: _defaultBoxSize[1] },
        boxSize = getBoxSize(markerConfig, labelConfig, align, scales.size, globalFontSize),
        hoverMarkerConfig = disabledOptions.marker,
        mergedHoverMarkerConfig = isEmptyObject(hoverMarkerConfig) ? hoverMarkerConfig : mergeJSON(hoverMarkerConfig, markerConfig),
        disabledLabelConfig = disabledOptions.label,
        isDisabled = (d, i) =>
            config.filter &&
            config.filter.range && // ZC2406
            userRanges &&
            userRanges.length &&
            userRanges[getListIndex(i, { config, ticks: ticks || data })].disabled;

    let { hAlign } = markerConfig,
        { width, height, size, overflow, totalHeight } = boxSize,
        { keepAligned = true, padding = 0, showAs } = labelConfig,
        isHorizontal = align.layout === HORIZONTAL_ALIGN,
        markerEle,
        markerOnEnd = ele => {
            // TODO: clean
            ele.each(function(d, i) {
                if (isDisabled(d, i)) {
                    highlightMarker(d3_select(this), boxSize, markerConfig, mergedHoverMarkerConfig);
                }
            });
            markerEle = ele;
        },
        labelOnEnd = (ele) => {
            ele.each(function(d, i) {
                // TODO: clean
                if (isDisabled(d, i)) {
                    highlightLabel(d3_select(this), config, globalFont, disabledLabelConfig);
                }
            });

            if (isHorizontal && !computedInfo) {//avoid returing incase of not size legend
                return;
            }

            let labelNodes = ele.nodes(),
                margin_top = function(d, i, ele){
                    var labelHeight = labelNodes[i].offsetHeight;
                    return getColorBoxVerticalPosition(
                        markerConfig ? markerConfig.vAlign : CENTER_ALIGN,
                        labelHeight,
                        boxSize.height(d),
                        globalFontSize,
                        scales.size ? boxSize.height(d) : markerConfig.size,
                        labelConfig.textOverflow
                    );
                }
            
            markerEle.style('margin-top', function(d, i){
                if(unitTestObj){
                    unitTestObj.legend_grp.legend_nodes.colorBox.margin_top = margin_top(d, i);
                }
                return `${margin_top(d, i)}px`;
            })
            ele.attr('title',function(d, i) {
                var title = labelConfig.tooltip ? this.textContent : null;
                if (unitTestObj) {
                    // unitTestObj.legend_grp.legend_nodes.label.title[i] = title;
                    unitTestObj.legend_grp.legend_nodes.label.title = title;
                    // the commented line must come in her, inorder to match with old config did like this
                }
                return title;
            })
        },
        boxStyles = {},
        labelStyles = {},
        liStyles = {};

    if (!isBoxEnabled(markerConfig)) {
        padding = 0;
    }

    if (keepAligned) {
        let boxMargin = NULL,
            labelMargin = NULL,
            liMargin = NULL,
            zIndexLI = NULL,
            lineHeight;

        if (isHorizontal) {
            boxMargin = d => `${(mathMax(globalFontSize, bound.height) - height(d)) / 2}px`;
            labelMargin = () => `${padding}px`;
            lineHeight = `${bound.height}px`;
            // boxMargin = d =>
            //     `${vAlign === BOTTOM_ALIGN ? bound.height - size(d)[1] : vAlign === CENTER_ALIGN ? (bound.height - size(d)[1]) / 2 : 0}px`;
            // labelMargin = d =>
            //     `${(vAlign === BOTTOM_ALIGN ? 0 : vAlign === CENTER_ALIGN ? (bound.height - size(d)[1]) / 2 : bound.height - size(d)[1]) +
            //         padding}px`;
        } else {
            boxMargin = d =>
                `${hAlign === RIGHT_ALIGN ? bound.width - width(d) : hAlign === CENTER_ALIGN ? (bound.width - width(d)) / 2 : 0}px`;
            labelMargin = d =>
                `${(hAlign === RIGHT_ALIGN ? 0 : hAlign === CENTER_ALIGN ? (bound.width - width(d)) / 2 : bound.width - width(d)) +
                    padding}px`;

            if (overflow === 'visible') {
                liMargin = (d, i) => {
                    let top = i ? -(height(d) - totalHeight) : 0;
                    return `${top}px 0 0 0`;
                };

                if (size(data[0])[0] < size(data[data.length - 1])[0]) {
                    zIndexLI = (d, i) => data.length - 1 - i;
                }
            }

            let textHeight = NULL;
            lineHeight = d => {
                if (!textHeight) {
                    textHeight = getTextHeight([labelConfig, globalFont, $ZCG]);
                }

                return `${mathMax(textHeight, height(d))}px`;
            };
        }

        boxStyles = {
            float: LEFT_ALIGN,
            display: 'inline-block',
            margin: d => (isHorizontal ? `${boxMargin(d)} 0 0 0` : `0 0 0 ${boxMargin(d)}`)
        };
        labelStyles = {
            display: showAs ? 'inline-block' : NULL,
            margin: d => `0 0 0 ${labelMargin(d)}`
        };

        liStyles = { 'line-height': lineHeight, margin: liMargin, 'z-index': zIndexLI };
    } else {
        if (isHorizontal) {
            let boxMargin = d => `${bound.height / 2 - height(d) / 2}px`;
            boxStyles = { margin: d => `${boxMargin(d)} auto 0px 0px` };
            boxStyles = {
                margin: d => {
                    let top = boxMargin(d);
                    return hAlign === 'left'
                        ? `${top} auto ${padding}px 0`
                        : hAlign === 'right'
                        ? `${top} 0 ${padding}px auto`
                        : `${top} auto ${padding}px auto`;
                }
            };
        } else {
            liStyles = { 'text-align': hAlign };
            boxStyles = {
                margin:
                    hAlign === 'right'
                        ? `0 0 ${padding}px auto`
                        : hAlign === CENTER_ALIGN
                        ? `0 auto ${padding}px auto`
                        : `0 auto ${padding}px 0`
            };
        }

        boxStyles.float = NULL;
        boxStyles.display = NULL;
        labelStyles.display = NULL;
        labelStyles.margin = NULL;
        liStyles = { ...liStyles, 'line-height': NULL, 'z-index': NULL, margin: NULL };
    }

    var markerInfo = computedInfo && computedInfo.marker,
        labelInfo = computedInfo && computedInfo.label,
        marker_styles = markerInfo && computedInfo.marker.styles,
        label_styles = labelInfo && computedInfo.label.styles;
    return {
        marker: { styles: { ...boxStyles, ...marker_styles }, onEnd: markerInfo && markerInfo.onEnd || markerOnEnd },
        label: { styles: { ...labelStyles, ...label_styles }, onEnd: labelInfo && labelInfo.onEnd || labelOnEnd },
        li: { styles: liStyles }
    };
}

export function getBoxSize(boxConfig, labelConfig, align, scale, fontSize) {
    let size = _size(scale, boxConfig, fontSize),
        overflow = getBoxOverfow(boxConfig, labelConfig, align),
        { width, height, shape } = boxConfig,
        totalWidth,
        totalHeight,
        clientWidth,
        clientHeight;

    if (overflow) {
        totalWidth = width;
        totalHeight = height;
    }

    if (overflow === 'hidden') {
        clientWidth = width;
        clientHeight = height;
    }

    return {
        width: d => clientWidth || size(d)[0],
        height: shape === 'half-circle' ? d => (clientHeight || size(d)[1]) / 2 : d => clientHeight || size(d)[1],
        totalWidth,
        totalHeight,
        overflow,
        size
    };
}

function getBoxOverfow(boxConfig, labelConfig, align) {
    let { width, height, overflow = 'hidden' } = boxConfig,
        isHorizontal = align.layout === HORIZONTAL_ALIGN;

    if (!(!isHorizontal && labelConfig.keepAligned) && overflow === 'visible') {
        return;
    }

    if (height || width) {
        return overflow;
    }
}

function _size(scale, boxConfig, fontSize) {
    if (scale) {
        return d => [scale(d), scale(d)];
    }

    return () => legend_helpers.getFilterBoxSize(boxConfig, fontSize);
}

function isBoxEnabled(boxConfig) {
    return boxConfig.show !== false;
}

function configByEncode(scales) {
    let _config = {};
    each(scales, (scale, encode) => (_config[encode] = { _type: 'encode', scale }));
    return _config;
}

export { ListParser };
