// @zohocharts/zc/src/view/list/list-helpers.js
//
import { TRANSPARENT, NONE, NULL } from '../../core/zc-constants';
import { HORIZONTAL_ALIGN } from '../../core/align-constants';
import { Registry } from '../../core/zc-registry';
import { invokeFunction, defined, pick } from '../../utils/zcutils';
import { legend_helpers } from '../../core/main-zc-init';
import { getHTML4SVGSymbol, getSVGSymbolString } from '../../utils/svg-symbols';

function markerDataURI(config, scale, advanced) {
    return d => {
        let width = scale.width ? scale.width(d) : scale(d),
            height = scale.height ? scale.height(d) : scale(d),
            dataURI = _getHTML4SVGSymbol(width, height, config);

        if (advanced) {
            return getSVGSymbolString(dataURI, width, height);
        } else {
            return dataURI;
        }
    };
}

function _getHTML4SVGSymbol(width, height, markerConfig) {
    let { fillColor = TRANSPARENT, strokeColor = TRANSPARENT, shape = 'square' } = markerConfig;

    return getHTML4SVGSymbol({...markerConfig, width: width, height: height, strokeColor: strokeColor, fillColor: fillColor, shape: shape}, true);
}

function markerBackground(boxConfig, boxSize) {
    let { fillColor = TRANSPARENT, strokeColor = TRANSPARENT, strokeWidth, borderRadius, shape = 'square' } = boxConfig,
        useSVGImage = shape !== 'square' && shape !== 'rect',
        border,
        background;

    if (useSVGImage) {
        background = markerDataURI(boxConfig, boxSize, true);
        border = NONE;
        borderRadius = NULL;
    } else {
        background = fillColor;
        border = (d, i) => `${pick(strokeWidth, 1)}px solid ${invokeFunction(strokeColor, d, i)}`;
        borderRadius = defined(borderRadius) ? `${borderRadius}px` : NULL;
    }

    return { background, border, ['border-radius']: borderRadius };
}

function getBoxSize(boxConfig, labelConfig, align, scale, fontSize) {
    let size = _size(scale, boxConfig, fontSize),
        overflow = getBoxOverfow(boxConfig, labelConfig, align),
        { width, height, shape } = boxConfig,
        totalWidth,
        totalHeight,
        clientWidth,
        clientHeight;

    if (overflow) {
        totalWidth = width;
        totalHeight = height;
    }

    if (overflow === 'hidden') {
        clientWidth = width;
        clientHeight = height;
    }

    return {
        width: d => clientWidth || size(d)[0],
        height: shape === 'half-circle' ? d => (clientHeight || size(d)[1]) / 2 : d => clientHeight || size(d)[1],
        totalWidth,
        totalHeight,
        overflow,
        size
    };
}

function getBoxOverfow(boxConfig, labelConfig, align) {
    let { width, height, overflow = 'hidden' } = boxConfig,
        isHorizontal = align.layout === HORIZONTAL_ALIGN;

    if (!(!isHorizontal && labelConfig.keepAligned) && overflow === 'visible') {
        return;
    }

    if (height || width) {
        return overflow;
    }
}

function _size(scale, boxConfig, fontSize) {
    if (scale) {
        return d => [scale(d), scale(d)];
    }

    return () => legend_helpers.getFilterBoxSize(boxConfig, fontSize);
}

export { markerDataURI, markerBackground, getBoxSize };
Registry.setComponent('listView.markerSize', getBoxSize);
