// @zohocharts/zc/src/scatter/parser/config/scatter-renderer-config.js
import { DataProcessor } from '../../../core/data-processor';
import { mergeJSON, pick, isUndefined, utils, validateNullVal } from '../../../utils/zcutils';
import { mathMax } from '../../../core/zc-constants';
// import { getStackGroupIndex } from '../../../core/data-processor/transforms/stack/stack-grouping';
import { isScatterChartType } from '../../../events/events-utils';
import { getChartTypeByIndex } from '../../../core/charttype-helpers';
import { chartcategory, charttype, stack } from '../../../core/main-zc-init';
import { isMultiColoring } from '../../../utils/colorutils';
import { getSegmentsForSymbol } from '../../../utils/svg-symbols';

export function getScatterCommonOption(conf, index, legendOrder, paneid) {
    legendOrder = pick(legendOrder, index);
    var rp = conf.params,
        stacked = rp.stacked,
        chartObj = rp.chartObj,
        xscale = rp.xscale,
        yscale = rp.yscale,
        hasNegValue = rp.hasNegValue,
        processor = chartObj.dataObject,
        commonPlotOptions = rp.commonPlotOptions,
        dataProcessorHelpers = DataProcessor.helpers,
        datasetIndex = chartObj.renderer[index].datasetIndex ? chartObj.renderer[index].datasetIndex() : 0,
        cType_Obj = getChartTypeByIndex(chartObj, paneid, legendOrder),
        cType = cType_Obj.name,
        cValue = cType_Obj.value,
        isBubble = chartcategory.bubble.has(cValue),
        scatterOrBubble = isScatterChartType(cValue),
        multiColoring = scatterOrBubble && isMultiColoring(commonPlotOptions, chartObj);

    var padding = (hasNegValue != null && xscale(hasNegValue) != null
        ? xscale(mathMax(hasNegValue, xscale.domain()[0]))
        : xscale.origRange[0]) + (xscale.bandwidth != null ? xscale.bandwidth() / 2 : 0);

    var addRangeBand = function () {
        return yscale.bandwidth ? yscale.bandwidth() / 2 : 0;
    };

    var stackGroupIndex;
    if (stacked) {
        var allChartTypes = processor.allChartTypes,
            stackGroup = processor.getStackGroup(),
            callback = d => d.some(seriesIndex => !chartObj.seriesdata[seriesIndex].disabled && isScatterChartType(allChartTypes[seriesIndex])),
            stackBarEncodes = stackGroup.filter(callback);

        stackGroupIndex = processor.categoryHierarchy ? 0 : stack.getStackGroupIndex(stackBarEncodes, index);
    } else {
        stackGroupIndex = 0;
    }

    var commonOption = {
        stacked: stacked,
        categories: rp.xscale.domain(),
        axisrotated: rp.axisrotated,
        bandwidth1: rp.xscale,
        bandwidth2: chartObj.barscale2,
        datasetIndex: datasetIndex,
        padding: padding,
        yscale: yscale,
        hasNegValue: rp.hasNegValue,
        yaxiscolumnorder: rp.yaxiscolumnorder,
        addRangeBand: addRangeBand,
        charttype: cType,
        isBubbleChart: isBubble,
        scatterOrBubble: scatterOrBubble,
        multiColoring: multiColoring,
        stackGroupIndex: stackGroupIndex
    };

    chartObj.renderer[index].commonOption = commonOption;
    return commonOption;
}

export function setScatterRendererConfig() {
    const chartObj = this.chart;
    const { yaxiscolumnorder, rendererIndex: index, paneid } = this.builderC;
    const rp = this.conf.params,
        dataset = rp.dataset,
        commonPlotOptions = rp.commonPlotOptions,
        plotoptions = this.data.plotoptions,
        { builderC, legendOrder, oldRenderer, data: currentdata } = this,
        _getXYvalue = builderC._getXYvalue;

    // predefined variables
    let color = rp.color;

    // data processor constants
    const dataProcessorHelpers = DataProcessor.helpers,
        isCatNullX = dataProcessorHelpers.isCatNull(dataset, 'x'),
        isCatNullY = dataProcessorHelpers.isCatNull(dataset, 'y', yaxiscolumnorder);

    // chart type constants
    const cType_Obj = getChartTypeByIndex(chartObj, paneid, index),
        cType = cType_Obj.value,
        chartName = cType_Obj.name,
        { isBubbleChart, scatterOrBubble, multiColoring } = this.commonOption,
        isPolar = chartObj.dataObject.isPolarAxisCategory;

    let markerOptions = (this.builderC.markerOptions = rp.markerOptions =
        (plotoptions && plotoptions.marker ?
            mergeJSON(commonPlotOptions.marker, plotoptions.marker, true)
            : commonPlotOptions.marker));

    let hoverOptions = (this.builderC.hoverOptions = rp.hoverOptions =
        plotoptions && plotoptions.hoveroptions
            ? mergeJSON(commonPlotOptions.hoveroptions, plotoptions.hoveroptions, true)
            : commonPlotOptions.hoveroptions);

    let symboltype = markerOptions.symbol,
        symbolsize = markerOptions.size,
        imageUrl = markerOptions.imageUrl,
        sizeRatio = markerOptions.sizeRatio ?? 0.5;

    if (!scatterOrBubble) {
        if (isUndefined(symbolsize)) {
            symbolsize =
                (plotoptions
                    ? utils.getValueOfPlotOptions(plotoptions, commonPlotOptions)('strokeWidth')
                    : commonPlotOptions.strokeWidth) || 2;
        }

        if (!imageUrl) {
            symbolsize *= isPolar ? 0.56 : 0.7;
        }
        color = currentdata.color;
    } else {
        symbolsize = validateNullVal(symbolsize, 1);
    }

    let outersize = markerOptions.outersize = symbolsize * 10,
        strokeWidth = validateNullVal(utils.getValueOfPlotOptions(plotoptions, commonPlotOptions)('strokeWidth'), 2);

    //Initiated in dataprocessor
    if (chartObj.dataObject.markerSize[chartName]) {
        const mSize = symboltype == 'line' || symboltype == 'vertical-line' ? strokeWidth || 1 : outersize;
        chartObj.dataObject.markerSize[chartName].add(mSize);
    }

    // renderer dependency functions
    function getSymbolSize(d, type = 'outer') {
        if (isPolar && (!d || d.isDummy)) {
            return 0;
        } else if (
            !isPolar &&
            cType !== charttype.stkdarea &&
            (isCatNullX(dataset.getX(d)) || isCatNullY(dataset.getY(d, yaxiscolumnorder)))
        ) {
            return 0;
        } else {
            let size = isBubbleChart ? _getXYvalue(d, 'z') : outersize;
            type !== 'outer' ? (size *= sizeRatio) : 0;
            return size;
        }
    }

    function getSymbolBBox(d) {
        var size = getSymbolSize(d);
        if (size > 0 && (symboltype == 'line' || symboltype == 'vertical-line')) {
            if (symboltype == 'line') {
                return {
                    width: size,
                    height: strokeWidth || 1
                };
            } else if (symboltype == 'vertical-line') {
                return {
                    height: strokeWidth || 1,
                    width: size
                };
            }
        } else {
            return null;
        }
    }
    //currently segments available only for line, vertical-line, whiskers symbol alone
    //for line - there is single segment
    //for whisker it is array of multiple line segments
    function getSegments(d) {
        var size = getSymbolSize(d);
        if (symboltype == 'line' || symboltype == 'vertical-line') {
            var _data = d._data();
            var symbolSegment = getSegmentsForSymbol(symboltype, size, _data._x, _data._y);
            if (symboltype == 'line' || symboltype == 'vertical-line') {
                return [symbolSegment];
            } else {
                return symbolSegment;
            }
        }
    }

    function pathData(d, i) {
        if (isPolar && (!d || d.isDummy || isCatNullX(dataset.getX(d)) || isCatNullY(dataset.getY(d, yaxiscolumnorder)))) {
            return [];
        } else if (
            !isPolar &&
            cType !== charttype.stkdarea &&
            (!d || d.isDummy || isCatNullX(dataset.getX(d)) || isCatNullY(dataset.getY(d, yaxiscolumnorder)))
        ) {
            return [];
        } else if (!isPolar && cType == charttype.stkdarea && (!d || d.isDummy)) {
            // to remove stacked area with dummy data added
            return [];
        } else {
            return [
                {
                    color: multiColoring ? chartObj.seriesColor[i % chartObj.seriesColor.length] : color,
                    index: multiColoring ? i : legendOrder,
                    data: d
                }
            ];
        }
    }

    // assign old renderer functions to new renderer 
    oldRenderer.getSymbolSize = getSymbolSize;
    oldRenderer.getSymbolBBox = getSymbolBBox;
    oldRenderer.getSegments = getSegments;
    oldRenderer.pathData = pathData;
}
