// @zohocharts/zc/src/behaviours/plot-brush/brushOptions.js
import { defaultTheme as options } from '../../core/defaulttheme';

options.brush = function(userProp) {
    return {
        enabled: false,
        type: 'x', // x, y, xy
        pan: true, // works only if zoom is enabled [* with shift key]
        snap: false, // snap for bar chart
        selectionArea: {
            fillColor: '#3a8aea',
            fillOpacity: 0.2,
            strokeColor: '#3a8aea',
            dashStyle: '',
            strokeWidth: 0.8
        },
        resetButton: {
            enabled: true,
            fillColor: '#3c3c3c',
            label: {
                marginTop: 6,
                marginRight: 10,
                marginBottom: 6,
                marginLeft: 10,
                fontSize: 14,
                fontColor: 'white',
                fontWeight: 500
            },
            border: {
                show: true,
                size: 1,
                radius: 2,
                color: 'transparent'
            }
        }
    };
};

export var _brushOptions = options.brush;
