// @zohocharts/zc/src/behaviours/plot-brush/plot-brush-arguments.js
import { Registry } from '../../core/zc-registry';
import { fixBrushDirection } from '../data-selection/chart/tools/selection-to-domain';

/**
 *
 * @param chart
 * @return {ZoomD3Options}
 */
export function _getPlotBrushArguments(chart) {
    let options = chart.systemConf.chart.brush,
        zoomOptions = chart.systemConf.chart.zoom,
        extent = [[0, 0], [chart.plotarea.width, chart.plotarea.height]],
        BrushD3Options = Registry.getComponent('BrushD3Options'),
        type = chart.isGeoMap ? 'xy' : fixBrushDirection(options.type, chart.axes.rotated),
        metaKey;

    // when brush and zoom enabled together and user wants to pan
    // we enable it on shift key and drag
    if (options.pan === true && zoomOptions.enabled) {
        metaKey = 'shiftKey';
    }

    return new BrushD3Options({
        type: type,
        extent: extent,
        element: chart.eventHandler.trackerGroup,
        metaKey,
        multitouch: options.multitouch,
        styles: {
            selection: {
                fillColor: 'transparent',
                strokeColor: 'transparent'
            }
        }
    });
}

Registry.setComponent('_getPlotBrushArguments', _getPlotBrushArguments);
