// @zohocharts/zc/src/scatter/parser/marks/scatter-inner-marks.js
import { MarksRegistry } from '../../../plot/MarksRegistry';
import { defined } from '../../../utils/zcutils';

export function _innerScatter(data) {
    let _this = this;
    let chart = this.chart,
        builderC = this.builderC,
        anime_config = builderC.animationOptions,
        racing = this.chart.racing && this.chart.racing.syncTransition,
        morphConfig = this.chart.systemConf.chart.plot.morph,
        morphing = morphConfig.enabled && this.chart.oldSeriesInfo,
        liveUpdate = builderC.liveUpdate,
        renderer = this.oldRenderer;

    const { markerOptions } = builderC,
        { imageUrl, innerStrokeWidth, innerFillOpacity, innerFillColor } = markerOptions,
        innerPathDisabled = imageUrl || (innerStrokeWidth === 0 && (innerFillOpacity === 0 || innerFillColor === 'transparent'));

    const KEY = 'inner',
        SUPER_KEY = 'scatter';

    let enterConfig, updateConfig;
    //* Add default enter, update and exit config here
    enterConfig = {
        fillOpacity: _this.helpers.styles[KEY].initialFillOpacity,
        strokeOpacity: _this.helpers.styles[KEY].initialStrokeOpacity,
        visibility: _this.helpers.styles[KEY].visibility
    };

    updateConfig = {
        d: function (d, i) { return _this.helpers.position[KEY].start.call(this, d) },
        fill: _this.helpers.styles[KEY].fill,
        stroke: _this.helpers.styles[KEY].stroke,
        strokeWidth: _this.helpers.styles[KEY].strokeWidth,
        dashStyle: _this.helpers.styles[KEY].strokeDasharray,
        lineCap: _this.helpers.styles[KEY].strokeLinecap,
        visibility: _this.helpers.styles[KEY].visibility
    };


    let enterMarks, updateMarks, exitMarks;
    //* Add default enter, update and exit marks here

    updateMarks = [
        {
            delay: anime_config._pathDelay,
            duration: anime_config._pathDuration,
            easing: anime_config.easingType,
            d: function (d, i) { return _this.helpers.position[KEY].end.call(this, d) },
            fillOpacity: _this.helpers.styles[KEY].fillOpacity,
            strokeOpacity: _this.helpers.styles[KEY].strokeOpacity
        }
    ];

    exitMarks = {
        duration: anime_config._exitduration
    };

    if (racing) {
        updateMarks = [
            {
                transition: racing,
                d: function (d, i) { return _this.helpers.position[KEY].end.call(this, d) },
                fillOpacity: _this.helpers.styles[KEY].fillOpacity,
                strokeOpacity: _this.helpers.styles[KEY].strokeOpacity
            }
        ];
        // TODO: Customize in racing cases
    } else if (morphing) {
        const morphDuration = morphConfig.duration;
        updateConfig = {
            strokeWidth: _this.helpers.styles[KEY].strokeWidth,
            fill: _this.helpers.styles[KEY].fill,
            stroke: _this.helpers.styles[KEY].stroke,
            dashStyle: _this.helpers.styles[KEY].strokeDasharray,
            lineCap: _this.helpers.styles[KEY].strokeLinecap,
            visibility: _this.helpers.styles[KEY].visibility
        };

        updateMarks = [
            {
                delay: morphDuration,
                d: function (d, i) { return _this.helpers.position[KEY].end.call(this, d) },
                fillOpacity: _this.helpers.styles[KEY].fillOpacity,
                strokeOpacity: _this.helpers.styles[KEY].strokeOpacity
            }
        ];
    } else if (liveUpdate) {
        updateMarks = [
            {
                duration: _this.phase == 'redraw' ? anime_config._duration : 0,
                d: function (d, i) { return _this.helpers.position[KEY].end.call(this, d) }
            }
        ];
    }

    let _innerScatterMark = {
        mark: 'point',
        selector: '.inner',
        data: innerPathDisabled ? [] : renderer.pathData,
        properties: {
            attrs: {
                class: (d, i) => `inner scatter-inner`
            }
        },
        onStart: function(ele){
            if(chart.racing && chart.racing.interrupt){
                ele.interrupt();
            }
        },
        config: {
            enter: enterConfig,
            update: updateConfig
        },
        anime: {
            // enter: enterMarks,
            update: updateMarks,
            exit: exitMarks
        }
    };
    return _innerScatterMark;
}

MarksRegistry.add('scatter.inner', _innerScatter);
