// @zohocharts/zc/src/scatter/parser/marks/scatter-outer-marks.js
import { MarksRegistry } from '../../../plot/MarksRegistry';
import { getRawColor } from '../../../utils/colorutils';
import { defined } from '../../../utils/zcutils';

export function _outerScatter(data) {
    let _this = this;
    let chart = this.chart,
        builderC = this.builderC,
        redraw = this.phase === 'redraw',
        anime_config = builderC.animationOptions,
        racing = this.chart.racing && this.chart.racing.syncTransition,
        morphConfig = this.chart.systemConf.chart.plot.morph,
        morphing = morphConfig.enabled && this.chart.oldSeriesInfo,
        liveUpdate = builderC.liveUpdate,
        index = builderC.rendererIndex,
        renderer = this.oldRenderer,
        seriesIdx = renderer.commonRendererProp().datasetIndex();

    const KEY = 'outer',
        SUPER_KEY = 'scatter';

    let enterConfig, updateConfig, exitConfig;
    //* Add default enter, update and exit config here
    if (!redraw) {
        enterConfig = {
            fillOpacity: _this.helpers.styles[KEY].initialFillOpacity,
            strokeOpacity: _this.helpers.styles[KEY].initialStrokeOpacity,
            visibility: _this.helpers.styles[KEY].visibility
        }
    }

    updateConfig = {
        d: function (d, i) { return _this.helpers.position[KEY].start.call(this, d) },
        strokeWidth: this.helpers.styles[KEY].strokeWidth,
        fill: this.helpers.styles[KEY].fill,
        stroke: this.helpers.styles[KEY].stroke,
        dashStyle: this.helpers.styles[KEY].strokeDasharray,
        lineCap: this.helpers.styles[KEY].strokeLinecap,
        visibility: _this.helpers.styles[KEY].visibility
    };

    let enterMarks, updateMarks, exitMarks;
    //* Add default enter, update and exit marks here

    updateMarks = [
        {
            delay: anime_config._pathDelay,
            duration: anime_config._pathDuration,
            easing: anime_config.easingType,
            d: function (d, i) { return _this.helpers.position[KEY].end.call(this, d) },
            fillOpacity: this.helpers.styles[KEY].fillOpacity,
            strokeOpacity: this.helpers.styles[KEY].strokeOpacity
        }
    ];

    exitMarks = {
        duration: anime_config._exitduration
    };

    //* Update marks for racing, morphing and liveUpdate charts
    if (racing) {
        updateMarks = [
            {
                transition: racing, 
                d: function (d, i) { return _this.helpers.position[KEY].end.call(this, d) },
                fillOpacity: this.helpers.styles[KEY].fillOpacity,
                strokeOpacity: this.helpers.styles[KEY].strokeOpacity
            }
        ];

        // TODO: Customize for racing cases
    } else if (morphing) {
        const oldseriesData = chart.oldSeriesInfo.get(seriesIdx),
            oldseriesColor = oldseriesData.color,
            morphDuration = morphConfig.duration,
            multiColoring = _this.commonOption.multiColoring;

        updateMarks = [
            {
                duration: morphDuration,
                easing: anime_config.easingType,
                fill: function (d, i) {
                    return multiColoring ? getRawColor(chart, d.index, d.data, seriesIdx, null, index)
                        : getRawColor(chart, 0, d.data, seriesIdx, null, index);
                },
                d: {
                    property: 'attrTween',
                    to: function (d, i) {
                        // TODO: Scatter morphing update marks
                        let interpolator = _this.helpers.position.scatter_morphing.interpolator[d.data.itemIndex];
                        return function (t) {
                            return _this.helpers.position.scatter_morphing.line_d3(
                                interpolator.map(function (p) {
                                    return p(t);
                                })
                            );
                        };
                    }
                }
            },
            {
                ...updateConfig,
                duration: 0
            }
        ];

        updateConfig = {
            fill: (d, i) => oldseriesColor[d.index],
        };

    } else if (liveUpdate) {
        updateMarks = [
            {
                duration: _this.phase == 'redraw' ? anime_config._duration : 0,
                d: function (d, i) { return _this.helpers.position[KEY].end.call(this, d) }
            }
        ];
    }

    let _outerScatterMark = {
        mark: 'point',
        selector: '.outer',
        data: renderer.pathData,
        properties: {
            attrs: {
                class: (d, i) => `outer scatter-outer`
            }
        },
        onStart: function(ele){
            if(chart.racing && chart.racing.interrupt){
                ele.interrupt();
            }
        },
        config: {
            enter: enterConfig,
            update: updateConfig
        },
        anime: {
            // enter: enterMarks,
            update: updateMarks,
            exit: exitMarks
        }
    };
    return _outerScatterMark;
}

MarksRegistry.add('scatter.outer', _outerScatter);
