// @zohocharts/zc/src/scatter/parser/marks/scatter-marks.js
import { POINTTWEEN_ANIMATION } from '../../../core/animation-constants';
import { d3_select } from '../../../core/d3-constants';
import { ORDINAL_DATATYPE } from '../../../core/datatype-constants';
import { MarksRegistry } from '../../../plot/MarksRegistry';
import { setOpacityForTickFilterElements } from '../../../plot/svg/set-pivot-attr';
import { blinkSVGsymbols } from '../../../utils/svg-symbols';

export function _mainScatter(data) {
    let _this = this;
    let chart = this.chart,
        builderC = this.builderC,
        markerOptions = builderC.markerOptions,
        anime_config = builderC.animationOptions,
        racing = this.chart.racing && this.chart.racing.syncTransition,
        morph = this.chart.systemConf.chart.plot.morph,
        morphing = morph.enabled && this.chart.oldSeriesInfo,
        liveUpdate = builderC.liveUpdate,
        attrs = {},
        renderer = this.oldRenderer,
        redraw = this.phase === 'redraw',
        imageUrl = markerOptions.imageUrl;


    const isOrdinalX = chart.dataset.getDataType('x', 0) === ORDINAL_DATATYPE;

    const KEY = 'scatter';

    let enterConfig, updateConfig, exitConfig;
    //* Add default enter, update and exit config here  - no animation purpose
    enterConfig = {
        transform: function (d, i) { return _this.helpers.position[KEY].start.call(this, { data: d, index: i }) }
    };

    updateConfig = {
        visibility: this.helpers.styles[KEY].visibility
    }

    if (redraw && imageUrl) {
        updateConfig.transform = function (d, i) { return _this.helpers.position[KEY].stop_1.call(this, { data: d, index: i }) };
    }

    let enterMarks, updateMarks, exitMarks;

    //* Default enter, update and exit marks - animation purpose
    enterMarks = {
        transition: racing,
        transform: function (d, i) { return _this.helpers.position[KEY].start.call(this, { data: d, index: i }) }
    };

    updateMarks = [
        {
            transition: racing,
            delay: anime_config._delay,
            duration: anime_config._duration,
            easing: anime_config.easingType,
            transform: function (d, i) { 
                return _this.helpers.position[KEY].end.call(this, { data: d, index: i }) }
        }
    ];

    exitMarks = {
        duration: anime_config._exitduration
    };

    //* Customization of config and marks for pointTween animation
    let onEnd_subseries;
    if (anime_config.type == POINTTWEEN_ANIMATION && !redraw) {
        updateMarks = null;
        onEnd_subseries = anime_config._intertween;
    }

    //* Update marks for racing, morphing and liveUpdate charts
    if (racing) {
        enterMarks = {
            duration: 0,
            transform: function (d, i) { 
                if(!_this.chart.racing.keyFramedRace){
                    //to go to the position directly, give duration 0 and give end position in start position.
                    return _this.helpers.position[KEY].end.call(this, { data: d, index: i })
                }
                return _this.helpers.position[KEY].start.call(this, { data: d, index: i }) 
            }
        };
        exitMarks = {
            transition: racing,
            duration: anime_config._exitduration,
            transform: function (d, i) { return _this.helpers.position[KEY].exit.call(this, { data: d, index: i }) }
        };
    }
    if (morphing) {
        enterConfig = {
            transform: function (d, i) { return _this.helpers.position[KEY].end.call(this, { data: d, index: i }) }
        };
        enterMarks = null;
    } else if (liveUpdate) {
        updateMarks = [
            {
                duration: _this.phase == 'redraw' ? anime_config._duration : 0,
                transform: function (d, i) { return _this.helpers.position[KEY].end.call(this, { data: d, index: i }) }
            }
        ];
    }

    if (this.chart.dataObject.hierarchical) {
        attrs = {
            levelbycolumn: d => d.levelByColumn,
            levelbyrow: d => d.levelByRow
        };
    }

    let _mainScatterPath = {
        mark: 'g',
        selector: '.subseries',
        data: function () {
            return data.disabled ? [] : data.data;
        },
        properties: {
            attrs: {
                class: (d, i) => {
                    d.subseriesIdx = i;
                    return `subseries subseries_${i}`;
                }
            }
        },
        anime: {
            exit: {
                delay: anime_config._exitduration
            }
        },
        onEnd: onEnd_subseries,     // for intertween animation call
        children: [
            {
                mark: "g",
                selector: '.scatter',
                data: function (d, i) {
                    return data.disabled ? [] : _this.helpers.filterData(d);
                },
                dataAccessor: racing && isOrdinalX ? _ => chart.dataset.getX(_) : null,
                config: {
                    enter: enterConfig,
                    update: updateConfig,
                    exit: exitConfig
                },
                anime: {
                    enter: enterMarks,
                    update: updateMarks,
                    exit: exitMarks
                },
                properties: {
                    attrs: {
                        ...attrs,
                        class: 'scatter zc_scatter'
                    },
                    styles: {
                        transition: 'opacity ease .2s'
                    }
                },
                onStart: function(ele){
                    if(chart.racing && chart.racing.interrupt){
                        ele.interrupt();
                    }
                },
                onEnd: function (ele) {
                    setOpacityForTickFilterElements(ele, chart, 'scatter');

                    blinkSVGsymbols(ele.selectAll(`.outer,.inner`), _this.builderC.blinkOptions, chart);

                    ele.attr("size", function (d) {
                        return renderer.getSymbolSize(d);   //Saving the size, makes redraw animation cool
                    })

                    if (morphing && chart.oldSeriesInfo) {
                        delete chart.oldSeriesInfo;
                    }
                },
                children: [
                    MarksRegistry.get('scatter.outer').call(this, data),
                    MarksRegistry.get('scatter.inner').call(this, data)
                ]
            }
        ]
    };
    return _mainScatterPath;
}

MarksRegistry.add('scatter.default', _mainScatter);
