// @zohocharts/zc/src/axis/radial/parser/encode/radial-majorticks-encode.js
import { translateString } from "../../../../utils/svg-transforms";
import { pick } from "../../../../utils/zcutils";

const KEY = 'gridline'; //* majorticks
export function radial_majorticks_position() {
    const self = this;
    const { renderer, helpers, component, config } = this,
        { position } = helpers,
        { radius, scale, xscale, gridPath } = component;

    // const majorTicks = renderer.majorTicks();

    position[KEY] = {
        transform: d => {
            return translateString([0, radius - scale(d)]);
        },
        d: d => {
            return gridPath(radius - scale(d), xscale, config)
        }
    };

}

export function radial_majorticks_styles() {
    const { renderer, helpers, config } = this,
        { styles } = helpers,
        { grid } = config;;

    const strokeWidth = pick(grid.strokeWidth, 1);

    styles[KEY] = {
        fill: 'none',
        // shapeRendering: 'crispedges',
        stroke: grid.color,
        strokeWidth: strokeWidth
    };
}
