// @zohocharts/zc/src/axis/radial/parser/encode/radial-ticklabel-encode.js
import { $ZCG } from "../../../../core/zcglobal";
import { getValByPriority, invokeFunction, utils, validateNullVal } from "../../../../utils/zcutils";

const KEY = 'ticklabel';
export function radial_ticklabel_position() {
    const self = this;
    const { renderer, helpers, commonOption, scale: scale1, chart, data, axis } = this,
        { tickTransform } = data,
        { position } = helpers,
        tickObj = axis.ticks.tickValueObj,
        { rotationInfo, tickConstants, spacing } = commonOption,
        { x, y, l } = tickConstants;

    function ticklabelPosition(d, attr, val, callback = tickTransform) {
        if (callback) {
            let customTickTransform = invokeFunction(callback, d, scale1, chart);
            val += attr == 'x' ? customTickTransform[0] : customTickTransform[1];
        }
        return val;
    }

    position[KEY] = {
        [x]: d => {
            return ticklabelPosition(d, x, l * spacing);
        },
        [y]: d => {
            return ticklabelPosition(d, y, 0);
        },
        dy: function (d, i) {
            return tickObj && tickObj.mode == 'zigzag' && i % 2 && rotationInfo.dy1 ? rotationInfo.dy1 : rotationInfo.dy;
        },
        transform: d => {
            let t = [0, 0];
            if (tickTransform) {
                t = invokeFunction(tickTransform, d, scale1, chart);
            }
            return `rotate(${rotationInfo.rotate[0]} ${rotationInfo.rotate[1] + t[0]} ${rotationInfo.rotate[2] + t[1]})`;
        }
    };
}

export function radial_ticklabel_styles() {
    const { renderer, helpers, config, commonOption, chart } = this,
        { styles } = helpers,
        { rotationInfo, tickConstants, spacing } = commonOption,
        { ticklabel: ticklabeldata } = config,
        events_tick = ticklabeldata.events;

    const hierarchy = [ticklabeldata, chart.systemConf.chart, chart.systemConf.canvas, $ZCG],
        fSize = utils.getVal(getValByPriority(hierarchy, 'fontSize')),
        fontFamily = getValByPriority(hierarchy, 'fontFamily'),
        fontWeight = getValByPriority(hierarchy, 'fontWeight'),
        fontStyle = getValByPriority(hierarchy, 'fontStyle'),
        textShadow = getValByPriority(hierarchy, 'textShadow'),
        textDecoration = getValByPriority(hierarchy, 'textDecoration'),
        fill = getValByPriority(hierarchy, 'fontColor'),
        eventDisabled = events_tick != null && validateNullVal(events_tick.enabled, true) == false,
        cursor = eventDisabled ? 'auto' : events_tick ? events_tick.cursor || 'pointer' : 'pointer',
        textAnchor = rotationInfo.textAnchor,
        strokeColor = getValByPriority(hierarchy, 'strokeColor');

    const { x, l } = tickConstants;

    styles[KEY] = {
        fontSize: `${fSize}px`,
        fontFamily, fontWeight, fontStyle, textShadow,
        textDecoration, fill, cursor, textAnchor,
        [x]: l * spacing
    }
}
