// @zohocharts/zc/src/formatter/formatter.js
//$Id$
//formatter.js

import {
    invokeFunction,
    simpleClone,
    isArray,
    encodeHTML,
    decodeHTML,
    defined,
    validateNullVal,
    pInt,
    pick,
    toCapitalize
} from '../utils/zcutils';
import { NUMERIC_DATATYPE, TIME_DATATYPE } from '../core/datatype-constants';
import { STACKLABELS_LABELTYPE, DATALABELS_LABELTYPE } from '../label/labeltype-constants';
import { mathFloor, EPSILON } from '../core/zc-constants';
import { $Browser } from '../core/browser-detect';
import { d3_format, d3_timeFormat, d3_timeFormatUTC, d3_timeParse } from '../core/d3-constants';
import { d3_formatLocale, d3_timeFormatLocale } from '../core/d3-constants';
import { getLocale } from './_locale';
import { Registry } from '../core/zc-registry';

export function getFormattedValue(colInfo, value, dataType, chartObj, type, toDecode, arraySeperator, levelIndex) {
    if (!dataType) {
        return encodeOrDecode(value);
    }
    if (value == null || (value.constructor == Array && value[0] === null)) {
        return colInfo && colInfo.nullformatting != null
            ? encodeOrDecode(invokeFunction(colInfo.nullformatting, value, dataType, colInfo, type, chartObj))
            : value;
    }
    var format = colInfo ? colInfo[`${type}format`] || colInfo.format : null;
    if (!format || !Object.keys(format).length) {
        format = getDefaultFormat(dataType, colInfo);
    }
    return value.constructor == Array //For partition charts, array of values will come here to format
        ? value
              .map(function(d) {
                  return applyFormat(d);
              })
              .join(defined(format.arraySeperator) ? format.arraySeperator : arraySeperator)
        : applyFormat(value);

    function applyFormat(value) {
        if (dataType === TIME_DATATYPE && !format.customFormat) {
            value = generateDateVal(value, format, colInfo);
        } else if (
            dataType === NUMERIC_DATATYPE &&
            (type === 'axis' || type === STACKLABELS_LABELTYPE || (type === DATALABELS_LABELTYPE && chartObj.dataObject.isHierarchy)) &&
            typeof value === 'number'
        ) {
            if (countDecimals(value) > 12) {
                //Test_895.json
                value = parseFloat(value.toFixed(12)); //To avoid javascript floating point issue in axis ticks Test_559.json
            }
        }

        if (levelIndex && format.customFormat) {
            colInfo = simpleClone(colInfo);
            colInfo.levelindex = levelIndex;
        }
        var formattedValue =
            typeof format === 'function'
                ? format(value, dataType, colInfo)
                : format.customFormat
                ? invokeFunction(format.customFormat, value, dataType, colInfo, type, chartObj)
                : generateFormatSpecifier(format, value, colInfo, dataType, levelIndex, type, chartObj);
        return encodeOrDecode(formattedValue);
    }

    function encodeOrDecode(val) {
        return toDecode != false ? decodeHTML(val) : encodeHTML(val);
    }
}

function getDefaultFormat(type, colInfo) {
    var defaultFormat = {};
    switch (type) {
        case 'time':
            if (colInfo && colInfo.groupby) {
                defaultFormat = {};
            } else if (colInfo && colInfo.subfunction) {
                defaultFormat = getTimeSubfunctionFormatter(colInfo.subfunction, colInfo.useUTC, colInfo.locale);
            } else {
                defaultFormat.specifier = 'YYYY';
            }

            break;
        case 'numeric':
            defaultFormat.thousandSeperator = ',';
            break;
        case 'geo': {
            // nothing yet
            break;
        }
        case 'percent':
            defaultFormat.decimalPlaces = 1;
            defaultFormat.suffix = '%';
            defaultFormat.trailingZeros = false;
    }
    return defaultFormat;
}

export function generateFormatSpecifier(format, value, colInfo, dataType, levelIndex, type, chartObj) {
    let { locale } = colInfo || {};
    var formattedValue;
    switch (dataType) {
        case 'time': {
            // if useUTC is true use utc time formatter else use local time formatter
            var d3Format = colInfo && colInfo.useUTC === true ? getFormatter('time', locale, 'utcFormat') : getFormatter('time', locale);
            var specifier = format.specifier,
                groupOrder;

            if (defined(colInfo)) {
                groupOrder = isArray(colInfo.groupby) ? colInfo.groupby : colInfo.AUTO_GROUP_ORDER; //#ZC1015
            }

            if (isArray(specifier) || isArray(groupOrder)) {
                //For multiple line lables with the given array of specifier
                formattedValue = [];
                specifier = specifier || [null];
                if (defined(levelIndex)) {
                    specifier = defined(specifier) ? [specifier[levelIndex]] : [null];
                }
                specifier = isArray(specifier) ? specifier : [specifier];
                // specifier = (defined(levelIndex)) ? [specifier[levelIndex]] : specifier; //for timegroupby with specifier index
                specifier.forEach(function(d, i) {
                    var datetimespecifier = generated3TimeSpecifier(d),
                        formatter = generateTimeFormatter(d3Format, datetimespecifier, colInfo, levelIndex);
                    formattedValue[i] = formatter ? formatter(value) : value;
                });
                if (type == DATALABELS_LABELTYPE) {
                    formattedValue = formattedValue && formattedValue.toString();
                }
            } else if (!specifier) {
                var dateTimeSpecifier = getDefaultFormat(dataType, colInfo);
                formattedValue = dateTimeSpecifier(value);
            } else {
                var datetimespecifier = generated3TimeSpecifier(specifier);
                if (specifier === 'auto' || datetimespecifier == null) {
                    formattedValue = value;
                } else {
                    if (CUSTOM_SPECIFIER_REGEX.test(datetimespecifier)) {
                        var formatter = generateTimeFormatter(d3Format, datetimespecifier, colInfo, levelIndex);
                        formattedValue = formatter(value);
                    } else {
                        formattedValue = d3Format(datetimespecifier)(value);
                    }
                }
            }
            break;
        }
        case 'ordinal': {
            var lc = format.letterCase;
            value += ''; //toString
            formattedValue = lc
                ? lc === 'uppercase'
                    ? value.toUpperCase()
                    : lc === 'lowercase'
                    ? value.toLowerCase()
                    : toCapitalize(value)
                : value;
            break;
        }
        case 'geo': {
            var mapWith = format.mapBy,
                lc = format.letterCase;
            formattedValue = value;

            if (mapWith) {
                var feature = chartObj.dataObject.userMapFeatures.get(value);

                if (feature && feature.properties && feature.properties[mapWith]) {
                    formattedValue = feature.properties[mapWith];
                }
            }
            formattedValue = lc
                ? lc === 'uppercase'
                    ? formattedValue.toUpperCase()
                    : lc === 'lowercase'
                    ? formattedValue.toLowerCase()
                    : formattedValue[0].toUpperCase() + formattedValue.substring(1).toLowerCase()
                : formattedValue;
            break;
        }
        default: {
            if (format.type == 'decimal' && pInt(value) != value) {
                return ''; // Test_558.json
            }

            var decimalPlaces = format.decimalPlaces;
            value = format.type == 'decimal' || decimalPlaces === 0 ? parseFloat(value) : value; //Decimal places = 0
            var currency = format.currency;
            var numericspecifier = generated3NumericSpecifier(format, null, locale, value),
                toRemoveTrailingZeros = false;
            formattedValue = numericspecifier(value);

            if (/(e|s)/gi.test(getd3NumericFormatterType(format))) {
                if (defined(decimalPlaces)) {
                    formattedValue = formattedValue.replace(/[^a-z]+/i, function() {
                        return generated3NumericSpecifier(format, `.${decimalPlaces}f`, locale)(arguments[0]);
                    });
                } else {
                    toRemoveTrailingZeros = true;
                }
            } else if (defined(decimalPlaces) && decimalPlaces > 0) {
                formattedValue = formattedValue.replace(/[^a-z]+/i, function() {
                    var number = +arguments[0].replace(/,/g, ''),
                        fixedNumber = toFixedNumber(number, decimalPlaces);

                    return generated3NumericSpecifier(format, `.${decimalPlaces}f`, locale)(fixedNumber);
                });
            }

            if (format.signEnabled === false) {
                //1) null -> shows sign for -ve only 2) true -> shows sign for +ve and -ve 3) false -> no sign shown for both +ve and -ve
                formattedValue = formattedValue.substring(1, formattedValue.length);
            }

            if (!validateNullVal(format.trailingZeros, true) || toRemoveTrailingZeros) {
                // Remove trailing zeros
                var unitSuffix = '';
                formattedValue = formattedValue.replace(/[a-z]+(\+|-)*(\d)*$/i, function() {
                    unitSuffix = arguments[0];
                    return '';
                });

                formattedValue =
                    formattedValue.replace(/\.(\d)*0*$/gi, function() {
                        var matchedString = arguments[0];
                        return !parseFloat(matchedString)
                            ? ''
                            : `.${
                                  exponentialToDecimal(+matchedString)
                                      .toString()
                                      .split('.')[1]
                              }`; //Test_822.json, ZC1418
                    }) + unitSuffix;
            }

            formattedValue = applySeperators(format, formattedValue);

            if (currency != null) {
                //Add the currency symbol
                formattedValue = /(\+|\-)/.test(formattedValue[0])
                    ? formattedValue[0] + currency + formattedValue.slice(1)
                    : currency + formattedValue;
            }
        }
    }

    //Add prefix or suffix
    formattedValue = addPrefixSuffix(format, formattedValue);

    if (format.breakSpace) {
        formattedValue = formattedValue.split(' ');
    }
    return formattedValue;
}

function getTimeSubfunctionFormatter(subfunctype, useUTC, locale) {
    useUTC = useUTC === true;
    var formatter = useUTC ? getFormatter('time', locale, 'utcFormat') : getFormatter('time', locale);
    switch (subfunctype) {
        case 'YEAR':
        case 'year':
            return formatter('%Y');
        case 'MONTH':
        case 'month':
            return formatter('%B');
        case 'ABSDAY':
            return formatter('%a');
        case 'DAY':
        case 'day':
            return formatter('%e');
        case 'WEEKDAY':
            return formatter('%A');
        case 'WEEK':
        case 'week':
            return formatter('Week %W');
        case 'ABSHOUR':
            return formatter('%I');
        case 'ABSSECOND':
            return formatter('%S');
        case 'ABSMINUTE':
        case 'MINUTE':
            return formatter('%M');
        case 'MILLISEC':
            return formatter('%L');
        case 'HOUR':
            return formatter('%H hrs');
        case 'DATETIME':
            return formatter('%d %b %Y %H:%M:%S');
        case 'DATE':
            return formatter('%x');
        case 'TIME':
            return formatter('%X');
        case 'TIMESTAMP':
            return formatter('%x %X hrs');
        case 'ABSMONTH':
            return formatter('%b %Y');
        case 'ABSWEEK':
            return formatter('W' + '%U' + ' ' + '%Y');
        case 'ABSQUARTER':
            return function(d) {
                var year = useUTC ? d.getUTCFullYear() : d.getFullYear();
                return formatter(`Q${getQuarter(d, useUTC)} ${year}`);
            };
        case 'QUARTER':
            return function(d) {
                return formatter(`Q${getQuarter(d, useUTC)}`);
            };
        case 100:
            return formatter('%d %b %Y %H:%M:%S %L');
        default:
            throw `UNKNOWN TYPE 111: ${subfunctype}`;
    }
}

export function generateDateVal(value, format, colInfo, toProcess) {
    //var dateValue = (format != null && format.specifier != null)? d3_timeFormat(format.specifier).parse(value):new Date(value.toString());
    var inputformat = colInfo ? colInfo.inputformat : null;
    var dateValue = new Date(value.toString() === '0' ? +value : value.toString()); // For 1970, the timestamp will be 0. new Date("0") will return 1999
    var customFunc =
        typeof inputformat === 'function' ? inputformat : typeof window[inputformat] === 'function' ? window[inputformat] : null;
    if (!(value instanceof Date) && inputformat != null && (customFunc || inputformat.toLowerCase() != 'timestamp')) {
        value = !customFunc && !(typeof value == String) ? value.toString() : value;
        var formatter = getFormatter('time', colInfo.locale, 'parse');
        var parsedVal = customFunc ? customFunc(value, colInfo) : formatter(generated3TimeSpecifier(inputformat))(value);
        dateValue = parsedVal == null ? dateValue : new Date(parsedVal);
    }
    if (isNaN(dateValue)) {
        dateValue = new Date(+value);
    }
    return dateValue;
}
export function getFormattedValueByPercent(colInfo, value, dataType, chartObj, type) {
    colInfo = colInfo || {};
    type = `${type && colInfo[`${type}percentformat`] ? type : ''}percent`;

    if (!colInfo[`${type}format`]) {
        colInfo.percentformat = getDefaultFormat('percent');
    }

    return getFormattedValue(colInfo, value, dataType, chartObj, type);
}

//Format Utils
var timeSpecifiers = {
        YYYY: '%Y',
        MMMM: '%B',
        EEEE: '%A',
        EEE: '%a',
        MMM: '%b',
        YY: '%y',
        MM: '%m',
        EE: '%U',
        dd: '%d',
        HH: '%H',
        hh: '%I',
        mm: '%M',
        ss: '%S',
        a: '%p',
        MILLI: '%L',
        z: '%Z',

        // ZC custom specifiers

        QT: '$Q'
    },
    numericSpecifiers = {
        decimal: 'd',
        exponent: 'e',
        float: 'f',
        binary: 'b',
        octal: 'o',
        hexadecimal: 'x',
        SIprefix: 's'
    };

/**
 * maps custom specifier with corresponding functions
 *
 * @type {{Q: getQuarter}}
 */
var CUSTOM_SPECIFIER_MAPPER = {
        Q: getQuarter
    },
    tempArrayKeys = Object.keys(CUSTOM_SPECIFIER_MAPPER).map(function(key) {
        return `\\$${key}`;
    }),
    CUSTOM_SPECIFIER_REGEX = new RegExp(tempArrayKeys.join('|'));

function generated3TimeSpecifier(specifier) {
    if (specifier) {
        for (var spec in timeSpecifiers) {
            var index = specifier.indexOf(spec);
            if (index >= 0 && specifier[index - 1] !== '%') {
                specifier = specifier.replace(new RegExp(spec, 'g'), timeSpecifiers[spec]);
            }
        }
    }
    return specifier;
}

/**
 *
 * @param {Function} d3Format
 * @param {String} specifier
 * @param {Boolean} useUTC
 * @returns {Function}
 */
function getSpecifierFormatter(d3Format, specifier, useUTC) {
    return function(date) {
        var val = d3Format(date); // calling d3 function

        var string = [],
            i = -1,
            j = 0,
            n = val.length,
            c;

        // search for custom specifier
        // below code taken from d3/d3-time-format

        while (++i < n) {
            if (val.charCodeAt(i) === 36) {
                // searches for zc custom specifier ($)
                string.push(val.slice(j, i));
                c = val.charAt(++i);
                c = defined(CUSTOM_SPECIFIER_MAPPER[c]) ? CUSTOM_SPECIFIER_MAPPER[c](date, useUTC) : c;
                string.push(c);
                j = i + 1;
            }
        }

        string.push(val.slice(j, i));
        return string.join('');
    };
}

function generateTimeFormatter(d3TimeFormat, specifier, colInfo, levelIndex) {
    var formatter;
    if (specifier) {
        formatter = getSpecifierFormatter(d3TimeFormat(specifier), specifier, colInfo.useUTC);
    } else if (colInfo.groupby) {
        //#ZC1015
        var groupOrder = isArray(colInfo.groupby) ? colInfo.groupby : colInfo.AUTO_GROUP_ORDER;
        var subFuncType = levelIndex === 0 || !defined(levelIndex) ? colInfo.subfunction : groupOrder[levelIndex - 1];
        formatter = getTimeSubfunctionFormatter(subFuncType, colInfo.useUTC, colInfo.locale);
    }
    return formatter;
}

function generated3NumericSpecifier(format, defaultSpecifier, locale, value) {
    var sign = defined(format.signEnabled) ? '+' : '',
        zeroPad = format.zeroPadding != null && format.zeroPadding.enabled ? `0${format.zeroPadding.digits}` : '',
        thousandSep = format.thousandSeperator != null ? ',' : '',
        type = getd3NumericFormatterType(format),
        isScientificFormat = /(e|s)/gi.test(type),
        decimalPlaces = format.decimalPlaces,
        decPlaces;

    if (decimalPlaces != null && decimalPlaces >= 0) {
        decPlaces = isScientificFormat ? ($Browser.IE ? '.20' : '') : `.${decimalPlaces === 0 ? decimalPlaces : decimalPlaces + 3}`; // #ZC898
    } else {
        decPlaces = isScientificFormat ? '' : `.${countDecimals(value)}`; //Test_67.json
    }

    var formatSpecifier = sign + zeroPad + thousandSep + (defaultSpecifier || decPlaces + type);
    return getFormatter(undefined, locale)(formatSpecifier);
}

function getd3NumericFormatterType(format) {
    var decimalPlaces = format.decimalPlaces,
        type = format.type;

    return numericSpecifiers[type] || 'f';
}

function applySeperators(format, formattedValue) {
    var thSep = format.thousandSeperator,
        deSep = format.decimalSeperator,
        split = formattedValue.split(/\./);
    return (
        (thSep != null && thSep != ',' ? split[0].replace(/,/g, thSep) : split[0]) +
        (!split[1] ? '' : (deSep != null ? deSep : '.') + split[1])
    );
}

/**
 * Adds prefix and suffix
 * @param  {Object} format
 * @param  {String} formattedValue
 * @return {String}
 */
function addPrefixSuffix(format, formattedValue) {
    if (isArray(formattedValue)) {
        return formattedValue;
    }

    var prefix = pick(format.prefix, ''),
        suffix = pick(format.suffix, '');

    return prefix + formattedValue + suffix;
}

function countDecimals(value) {
    if (isNaN(value)) {
        return 0;
    } else {
        value = +value; //Important: if value is string (e.g: "1.0000000455121E12") throws error
        if (Math.floor(value) == value) {
            return 0;
        }
    }

    var convertedNumber = exponentialToDecimal(value);
    return convertedNumber.toString().split('.')[1].length || 0;
}

function exponentialToDecimal(num) {
    //https://gist.github.com/jiggzson/b5f489af9ad931e3d186
    //if the number is in scientific notation remove it
    if (/\d+\.?\d*e[+-]*\d+/i.test(num)) {
        var zero = '0',
            parts = String(num)
                .toLowerCase()
                .split('e'), //split into coeff and exponent
            e = parts.pop(), //store the exponential part
            l = Math.abs(e), //get the number of zeros
            sign = e / l,
            coeff_array = parts[0].split('.');
        if (sign === -1) {
            num = `${zero}.${new Array(l).join(zero)}${coeff_array.join('')}`;
        } else {
            var dec = coeff_array[1];
            if (dec) {
                l = l - dec.length;
            }
            num = coeff_array.join('') + new Array(l + 1).join(zero);
        }
    }
    return num;
}

/**
 * Returns absolute years, months, days till seconds from timestamp
 * @param {Number} ms date timestamp
 * @returns {{seconds: number, minute: number, hour: number, day: number, month: number, years: number}}
 */

export function millisecondToAbsoluteTime(ms) {
    var seconds = mathFloor(ms / 1000),
        minute = mathFloor(seconds / 60),
        hour = mathFloor(minute / 60),
        day = mathFloor(hour / 24),
        month = mathFloor(day / 30.436854166666667),
        years = mathFloor(day / 365.24225);

    return {
        seconds: seconds,
        minute: minute,
        hour: hour,
        day: day,
        month: month,
        years: years
    };
}

/**
 *
 * @param milliseconds
 * @returns {{subFunction: string, specifier: *}}
 */

export function getAutoTimeSubFunction(milliseconds) {
    var absoluteTime = millisecondToAbsoluteTime(milliseconds),
        subFunction = 'YEAR';

    if (absoluteTime.years > 1) {
        subFunction = 'YEAR';
    } else if (absoluteTime.month > 1) {
        subFunction = 'ABSMONTH';
    } else if (absoluteTime.day > 1) {
        subFunction = 'ABSDAY';
    } else if (absoluteTime.hour > 1) {
        subFunction = 'ABSHOUR';
    } else if (absoluteTime.minute > 1) {
        subFunction = 'ABSMINUTE';
    } else {
        subFunction = 'ABSSECOND';
    }

    return {
        subFunction: subFunction
    };
}

/**
 * (2.55).toFixed(1) returns 2.5 insteadof 2.6
 * stackoverflow.com/questions/49629597/js-decimal-precision-tofixed-not-returning-correct-value
 * stackoverflow.com/questions/11832914/round-to-at-most-2-decimal-places-only-if-necessary
 * @param  {Number} number
 * @param  {Number} decimal
 * @return {Number}
 */
export function toFixedNumber(number, decimal) {
    var tenToD = Math.pow(10, decimal);
    return +(Math.round((number + EPSILON) * tenToD) / tenToD).toFixed(decimal);
}

/**
 * it returns the quarter of the given date
 *
 * @param {Date} date
 * @param {useUTC} useUTC
 * @returns {Number}
 */
export function getQuarter(date, useUTC) {
    return mathFloor((useUTC ? date.getUTCMonth() : date.getMonth()) / 3) + 1;
}

let timeApiMap = { format: d3_timeFormat, parse: d3_timeParse, utcFormat: d3_timeFormatUTC };
function getFormatter(dataType = 'numeric', localeName, name = 'format') {
    let locale = getLocale(localeName);

    if (dataType === 'time') {
        return locale ? d3_timeFormatLocale(locale)[name] : timeApiMap[name];
    } else {
        return locale ? d3_formatLocale(locale)[name] : d3_format;
    }
}
