// @zohocharts/zc/src/legend/events/visual-effects/highlight-tooltip.js
import { LEFT_ALIGN, RIGHT_ALIGN, BOTTOM_ALIGN, TOP_ALIGN } from '../../../core/align-constants';
import { SVGTooltip } from '../../../tooltip/svg-tooltip';
import { mergeJSON, bbox } from '../../../utils/zcutils';
import { brushTooltipOptions } from '../../legend-brush-options';

export function HighlightTooltip(options) {
    let { config, orient, chart, element } = options,
        { systemConf } = chart,
        defaultOptions = {
            ...brushTooltipOptions(),
            fontSize: chart.globalProp.fontL1Size.l5,
            orient
        },
        tooltipOptions = {
            config: mergeJSON(defaultOptions, config, true),
            container: element,
            globalFont: [systemConf.chart, systemConf.canvas]
        };

    return new SVGTooltip(tooltipOptions);
}

export function getSVGTooltipPosition(ele, config, isHorizontal = true, translate = [0, 0]) {
    let { padding, margin, orient } = config,
        bb = bbox(ele.node()),
        bb_x,
        bb_y,
        bb_width,
        bb_height,
        outerSpace = padding + margin,
        x,
        y;

    if (isHorizontal) {
        bb_x = bb.x;
        bb_y = bb.y;
        bb_width = bb.width;
        bb_height = bb.height;
    } else {
        bb_x = bb.y;
        bb_y = bb.x;
        bb_width = bb.height;
        bb_height = bb.width;
    }

    switch (orient) {
        case BOTTOM_ALIGN:
            x = 0;
            y = bb_y - outerSpace;
            break;
        case RIGHT_ALIGN:
            x = bb_x - outerSpace;
            y = 0;
            break;
        case LEFT_ALIGN:
            x = bb_x + bb_width + outerSpace;
            y = 0;
            break;
        case TOP_ALIGN:
            x = 0;
            y = bb_y + bb_height + outerSpace;
    }

    return [x + translate[0], y + translate[1]];
}
