// @zohocharts/zc/src/bubblepie/bubblepiechart.js
//$Id$
//bubblepiechart.js
import { svg_plot_renderer, renderer_helpers, chartcategory } from '../core/main-zc-init';
import { bubblepieUtils } from './bubblepie-utils';
import { getAnimationConf } from '../utils/animationutils';
import { BUBBLEPIE_CHARTNAME, SCATTER_CHARTNAME } from '../core/chartname-constants';
import { mergeJSON, validateNullVal, arrayIncludes } from '../utils/zcutils';
import { appendEle } from '../utils/zcsvgutils';
import { hasColorColumnInNonAxis } from '../utils/colorutils';
import { NULL, mathMin } from '../core/zc-constants';
import { d3_select } from '../core/d3-constants';
import { BubblepieHighlighter } from './bubblepie-highlight';
import { BubbleHighlighter } from '../scatter/bubbleHighlight';
import { DataProcessor } from '../core/data-processor';
import { isFacet } from '../axis/hierarchy-axis-utils';
import { getChartTypeByIndex } from '../core/charttype-helpers';

/**
 *
 * @returns {chart} bubblePie.
 */
/*
 PENDING :
 -------
 1) Total datalabel
 2) Annotation at centre / on arc?
 */
svg_plot_renderer.bubblepie = function() {
    var shiftInfo,
        commonRendererProp,
        redraw = false,
        subRenderers;

    function chart(selection) {
        selection.each(function(data) {
            //renderer config values
            var rp = commonRendererProp.params,
                chartObj = rp.chartObj,
                index = rp.datasetIndex,
                commonPlotOptions = rp.commonPlotOptions,
                axisrotated = rp.axisrotated,
                stacked = rp.stacked,
                xscale = rp.xscale,
                yscale = rp.yscale,
                yaxiscolumnorder = rp.yaxiscolumnorder,
                labelQueue = rp.labelQueue,
                renderingOrder = rp.renderingOrderIndex,
                { bound, rendererIndex = index } = rp,
                container = d3_select(this),
                currentdata = data[index],
                data_bind = { data: currentdata, index: renderingOrder },
                plotoptions = currentdata.plotoptions;

            renderer_helpers.saveSeriesdataInRenderer(chartObj, index, rendererIndex);
            if (index !== chartObj.bubblePieRendererIndex) {
                chart.chartRendered = true;
                return;
            }

            subRenderers = [];
            chart.subrenderer = chart.subrenderer || [];
            if (!isFacet(chartObj)) {
                chartObj.subrenderer = chart.subrenderer;
            }

            var animationOptions = getAnimationConf(commonPlotOptions, plotoptions, chartObj, BUBBLEPIE_CHARTNAME);
            var transitionDuration = animationOptions.duration;
            commonRendererProp.params.hoverOptions =
                plotoptions && plotoptions.hoverOptions
                    ? mergeJSON(commonPlotOptions.hoveroptions, plotoptions.hoveroptions, true)
                    : commonPlotOptions.hoveroptions;

            var pieRenderer = chart.subrenderer, // TODO: check
                paneid = isFacet(chartObj) ? `${currentdata.row}_${currentdata.column}` : '0_0',
                chartTypeObj = getChartTypeByIndex(chartObj, paneid, index),
                isBubblepie = !arrayIncludes(chartcategory.nonaxis, chartTypeObj.value),
                dataIndex = isBubblepie ? validateNullVal(commonPlotOptions.dataindex, 3) : chartObj.dataset.yidx(yaxiscolumnorder),
                pieData = chartObj.dataObject.processBubblePieData(data, { dataIndex }),
                labelIndex = -1;

            if (chartObj.renderVisiblePointsOnly) {
                // #ZC2073
                let dataUpdate = DataProcessor.helpers.filterPointsInViewport(chartObj, pieData[0], 'bubblepie', 'xy', index);
                pieData = [dataUpdate];
            }

            var series = appendEle(container, 'g', pieData ? [data_bind] : [], '.', 'class', 'scatterseries');
            if (pieData) {
                pieData.forEach(function(piedatum, ssi) {
                    pieRenderer[ssi] = pieRenderer[ssi] || {};
                    var dataKeyFunc = function(d) {
                            var dt = d.reference;
                            return bubblepieUtils.getKey(NULL, NULL, chartObj, [d.x, d.y]); //#ZC1323
                        },
                        _data = bubblepieUtils.filterBySum(piedatum, dataIndex);

                    //ZC2763
                    _data = _data.filter(_d => {
                        let dt = _d.reference,
                            x = chartObj.dataObject.getXYvalue(axisrotated, stacked, xscale, yscale, yaxiscolumnorder, dt, 'x'),
                            y = chartObj.dataObject.getXYvalue(axisrotated, stacked, xscale, yscale, yaxiscolumnorder, dt, 'y');

                        return !isNaN(x) && !isNaN(y);
                    });

                    if (isFacet(chartObj)) {
                        _data.forEach(_d => {
                            _d.row = currentdata.row;
                            _d.column = currentdata.column;
                        });
                    }

                    var subseries = appendEle(series, 'g', [1], '.', 'class', `subseries_${ssi}`);
                    var _class = hasColorColumnInNonAxis(chartObj) ? 'zc_nonaxis' : 'scatter',
                        g = appendEle(subseries, 'g', _data, '.', 'class', _class, NULL, dataKeyFunc);

                    g.each(function(d, i) {
                        var dt = d.reference, //First reference data for the position
                            key = bubblepieUtils.getKey(NULL, NULL, chartObj, [d.x, d.y]),
                            x = chartObj.dataObject.getXYvalue(axisrotated, stacked, xscale, yscale, yaxiscolumnorder, dt, 'x') || 0,
                            y = chartObj.dataObject.getXYvalue(axisrotated, stacked, xscale, yscale, yaxiscolumnorder, dt, 'y') || 0,
                            size = isBubblepie
                                ? chartObj.dataObject.getXYvalue(
                                      axisrotated,
                                      stacked,
                                      xscale,
                                      yscale,
                                      yaxiscolumnorder,
                                      dt,
                                      'z',
                                      chartObj.zscale
                                  )
                                : mathMin(bound.width, bound.height, xscale.bandwidth ? xscale.bandwidth() : 1),
                            rendererName = isBubblepie ? 'pie' : chartTypeObj.name,
                            translate = isBubblepie
                                ? [x - size / 2, y - size / 2]
                                : chartObj.axes.rotated
                                ? [bound.width / 2 - size / 2, y - size / 2 - bound.y]
                                : [x - size / 2 - bound.x, bound.height / 2 - size / 2];

                        pieRenderer[ssi][key] = (pieRenderer[ssi][key] || svg_plot_renderer[rendererName]())
                            .commonRendererProp(commonRendererProp)
                            .plotarea({
                                width: size,
                                height: size,
                                left: x - size / 2,
                                top: y - size / 2,
                                translate
                            });
                        pieRenderer[ssi][key].chartRendered = false;
                        pieRenderer[ssi][key].subSeriesIndex = ++labelIndex;
                        subRenderers.push(pieRenderer[ssi][key]);

                        d3_select(this)
                            .attr('transform', `translate(${translate[0]},${translate[1]})`)
                            .call(pieRenderer[ssi][key]);
                    });
                });
            }

            function getSymbolSize(d) {
                return chartObj.dataObject.getXYvalue(axisrotated, stacked, xscale, yscale, yaxiscolumnorder, d, 'z', chartObj.zscale);
            }
            chart.getSymbolSize = getSymbolSize;

            if (isBubblepie) {
                chartObj.timerObj[chartObj.timerObj.length] = chart.renderedTimer = setTimeout(function() {
                    redraw = true;
                    chart.chartRendered = true;

                    var seriesData = chartObj.seriesdata;
                    seriesData.forEach(function(d, i) {
                        let chartType = getChartTypeByIndex(chartObj, paneid, i).value;
                        if (!chartcategory.bubblePie.has(chartType)) {
                            // For combinational chart
                            return;
                        }

                        var renderer = chartObj.renderer[i];
                        labelQueue.push({
                            renderer: renderer,
                            charttype: SCATTER_CHARTNAME,
                            chartObj: chartObj,
                            args: [d, i],
                            paneid: paneid
                        });
                        labelQueue.push({
                            renderer: renderer,
                            charttype: 'pieInnerLabels',
                            chartObj: chartObj,
                            args: [seriesData, i, pieData],
                            paneid: paneid
                        });
                    });
                }, transitionDuration);
            }
        });
        return chart;
    }
    chart.getHighlighter = function(chartObj, index) {
        if (!chart.highlighter) {
            chart.highlighter = new BubbleHighlighter(chartObj, chart);
        }
        return chart.highlighter;
    };
    chart.commonRendererProp = function(_) {
        if (!arguments.length) {
            return commonRendererProp;
        }
        commonRendererProp = _;
        return chart;
    };
    chart.shiftInfo = function(_) {
        if (!arguments.length) {
            return shiftInfo;
        }
        shiftInfo = _;
        return chart;
    };

    chart.endFunctions = function() {
        if (subRenderers.every(_ => _.chartRendered)) {
            chart.chartRendered = true;
        }
    };

    chart.destroy = function destroy () {
        subRenderers && subRenderers.forEach(renderer => renderer.destroy && renderer.destroy());
    }

    return chart;
};

/**
 *
 * @param chartInstance
 * @param seriesIndex
 * @param subSeriesIndex
 * @param point
 * @return {null|{start: *, end: *}}
 */
export function getBubblePieAngle(chartInstance, seriesIndex, subSeriesIndex, point) {
    var key = bubblepieUtils.getKey(point, seriesIndex, chartInstance),
        pieRenderer = chartInstance.subrenderer[subSeriesIndex][key];

    if (!pieRenderer) {
        //ZC1498
        return null;
    }

    return {
        start: pieRenderer.data[seriesIndex].startAngle,
        end: pieRenderer.data[seriesIndex].endAngle
    };
}

export var _bubblepieChart = svg_plot_renderer.bubblepie;
