// @zohocharts/zc/src/behaviours/plot-brush/chart/chart-brush-handler-mappings.js
import { HIDE_SELECTION_BOX, SHOW_SELECTION_BOX } from '../../visual-effects/chart/controllers/brush-selection-box';
import { ChartBrushListener } from '../../listeners/chart-brush-listener';
import { ChartBrushDataFinder } from '../../data-selection/chart/plot-brush-data-finder';
import { ApplyBrushSelectionOnPlot } from '../../visual-effects/chart/controllers/update-plot-by-selection';
import { ClearOldInfo_C } from '../../visual-effects/chart/controllers/clear-old-info';
import { positionResetButton, ShowBrushResetButton } from '../../visual-effects/chart/controllers/plot-reset-button';
import { Registry } from '../../../core/zc-registry';
import { d3_set } from '../../../core/d3-constants';
import { UpdateViewportArea_C } from '../../visual-effects/chart/controllers/update-viewport-area';
import { UPDATE_SCROLL_BY_DOMAIN } from '../../visual-effects/chart/controllers/scroll-domain';
import { $Browser } from '../../../core/browser-detect';
import { setBrushStatus } from '../../visual-effects/chart/controllers/set-brush-status';

let _start = [],
    _brush = [];

if ($Browser.isTouchDevice) {
    _brush = [ClearOldInfo_C];
} else {
    _start = [ClearOldInfo_C];
}

let _chartBrushHandlers = {
    LISTENER: {
        start: ChartBrushListener,
        brush: ChartBrushListener,
        end: ChartBrushListener
    },
    DATA_FINDER: {
        start: ChartBrushDataFinder,
        brush: ChartBrushDataFinder,
        end: ChartBrushDataFinder
    },
    BEHAVIOUR: {
        start: [..._start, SHOW_SELECTION_BOX],
        brush: [setBrushStatus, ..._brush],
        end: [HIDE_SELECTION_BOX, ApplyBrushSelectionOnPlot, UPDATE_SCROLL_BY_DOMAIN, ShowBrushResetButton, UpdateViewportArea_C]
    },
    LISTENER_TYPES: d3_set(['start', 'brush', 'end']),
    SUPPORTED: chart => chart.isAxisCategory && !chart.isPolarAxisCategory,
    UPDATE: function() {
        //ZC2307
        positionResetButton(this.eventManager.chart);
    }
};

Registry.setComponent('_chartBrushHandlers', _chartBrushHandlers);

export { _chartBrushHandlers };
