// @zohocharts/zc/src/polaraxis/parser/polaraxis-parser.js
import { ParserRegistry } from "../../core/parser-registry";
import { noOperation } from "../../core/zc-constants";
import { renderDom } from "../../mark/dom-render";
import { executeMarkOperators } from "../../plot/mark-operator-executor";
import { ZC_POLARAXIS_MARKS_MAPPING } from "./marks/polaraxis-mark-mapping";

export class PolarAxisParser {
    #container; #marks; #markMapping; #operators

    constructor(dependencies) {
        const { container, data, ticks, subticks, range, renderer, chart } = dependencies;
        this.data = data;
        this.ticks = ticks;
        this.subticks = subticks;
        this.range = range;
        this.renderer = renderer;
        this.chart = chart;
        this.#container = container;
        this.#initialize();
    }

    #initialize() {
        this.setMarkMapping();  // can be used later for custom marks mapping
        this.setMarkParser(['polar']);
    }

    container() {
        return this.#container
    }

    marks() {
        return this.#marks;
    }

    mapping() {
        return this.#markMapping;
    }
    operators() {
        return this.#operators;
    }

    #collectEncode() {
        // can be used later for collecting encodes
        this.operators().forEach(operator => {
            let config = (ParserRegistry.config(operator) || noOperation);
            config.call(this);

            let encode = (ParserRegistry.encodes(operator) || noOperation);
            encode.call(this);
        });
    }

    #allowedMarkOperator(operator) {
        return operator ? (this.#operators = operator, operator) : this.#operators;
    }

    setMarkMapping(mapping) {
        this.#markMapping = (mapping || ZC_POLARAXIS_MARKS_MAPPING).call(this);
    }

    setMarkParser(operator) {
        let operators = this.#allowedMarkOperator(operator),
            markMapping = this.#markMapping,
            customOperator = this.markOperator;

        this.#collectEncode();
        this.#marks = executeMarkOperators.call(this, this.data, markMapping, operators, customOperator);
    }

    render(parent) {
        parent = parent || this.container();
        this.#marks.map(d => {
            renderDom(d.mark, d.holder || parent)
        })
    }

    destroy() {
        this.#container = this.#marks = null;
    }
}
ParserRegistry.parser('PolarAxis', PolarAxisParser);
