// @zohocharts/zc/src/axis/ticks/axis-ticks.js
import { TIME_DATATYPE } from '../../core/datatype-constants';
import { hasAutoFormat } from '../tick-helpers';
import { defined } from '../../utils/zcutils';
import { getLongestTick } from '../tick-labels';
import { getTicksForAxis } from './tick-data';

export class AxisTicks {
    constructor(args, tickParser) {
        this.chart = args.chart;
        this.type = args.type;
        this.order = args.order;
        this.config = args.config;
        this.plotRect = args.plotRect;
        this.plotRect1 = args.plotRect1;
        this._order = args._order;
        this.orient = args.orient;
        this.datatype = args.datatype;
        this.column = args.column;

        this.initialize(args, tickParser);
    }
    initialize(_axisH, tickParser) {
        _axisH.axisInfo.logDomain = null;
        let tickH = (this.tickH = tickParser(_axisH));
        this.domain = tickH.domain();
        this.scale = tickH.scale(); //no need to execute here
        this.simpleScale = tickH.simpleScale();
        this.filter = tickH.filter();
        this.useTickAsIndices = tickH.useTickAsIndices;
        this.styles = tickH.styles();
        this.rotationConfig = tickH.rotation();
        this.rotation = this.rotationConfig.rotation;
        this.format = tickH.format();
        this.formatter = tickH.formatter();
        this.groupby = tickH.groupby();
        this.side = tickH.side();
        this.size = tickH.size();
        this.maxSize = tickH.maxSize(); //TODO check with maxttick
        this.margin = tickH.margin();
        this.padding = tickH.padding(this.margin);
        this.autoScrollX = tickH.autoScrollX();
        this.autoScrollY = tickH.autoScrollY();
        this.chartPadding = tickH.chartPadding();
        this.chartOuterPadding = tickH.chartOuterPadding();
        this.plotSize = tickH.plotSize();
        this.plotSize_no_zoom = tickH.plotSize_no_zoom();
        this.subFormatType = tickH.subFormatType;
        this.chartRect = tickH.chartRect();
        this.branchType = tickH.isBranchType && tickH.isBranchType();
        this.leafType = tickH.isLeafType && tickH.isLeafType();
        this.bzsEnabled = tickH.bzsEnabled();
        this.customTickInfo = {};
        this.isAutoFormat = this.datatype == TIME_DATATYPE ? hasAutoFormat(this.format) : null;
        this.formatAccessor = tickH.formatAccessor(_axisH, this);
    }
    collect() {
        this.tickInfo = getTicksForAxis.call(this);

        if (defined(this.customTickInfo.styles) || (this.format && this.format.format && this.format.format.breakSpace)) {
            this.tickInfo.longestTick = getLongestTick(
                this.ticks_formatted,
                this.ticks,
                this.styles.fontSize,
                this.styles.fontFamily,
                this.customTickInfo.styles
            );
        }
    }
}
