// @zohocharts/zc/src/axis/ticks/axis-group-precalculation.js
import { appendEle } from '../../utils/zcsvgutils';
import { renderDom } from '../../mark/dom-render';
import { NONE } from '../../core/zc-constants';
import { invokeFunction } from '../../utils/zcutils';

export function precalculateAxisGroup() {
    /**
     * Adding the axis label here for the only first time. Next time just calls the transform method.
     */

    let { type, order, chart, config, plotRect, viewParser, orient, rotated } = this,
        { dataObject, dataset, systemConf } = chart,
        tickMargin = this.axisTicks.margin,
        { hideTitle, titleConfig, titleMargin, branchTitleConfig } = this.axisTitle,
        { axisgroup } = this.axisTicks.precalculate,
        tick_data = this.axisTicks.tickInfo,
        tickPrecalculation = { ...this.axisTicks.precalculate, ...this.axisTicks.tickInfo },
        preCalcAxisLen;

    var isHorizontal = type === 'x' && !rotated,
        isVertical = !isHorizontal,
        axisLabelProperties = {
            systemConf: systemConf,
            bound: axisgroup,
            horizontal: isHorizontal,
            vertical: isVertical,
            axisOrder: order,
            plotRect: plotRect,
            hide: hideTitle,
            tickObj: tickPrecalculation,
            type: type,
            chartObj: chart
        },
        axislabelProp = { textAnchor: 'middle' };

    var axisSize = 0;
    var axisdata_merged = branchTitleConfig,
        axes = this.axisInfo,
        selector = type == 'x' ? `${type}-axis-${order}` : `${type}-axis-${order}`,
        axisG = chart.svg.selectAll(`g.${type}-axis`),
        axisWrap = appendEle(axisG, 'g', [config], '.', 'class', selector);

    let markExtender = this.axisView && this.axisView.conf.preMarkExtender,
        axisTitleMark = this.marks.axisTitle.onReady();
    if (markExtender) {
        axisTitleMark = invokeFunction(markExtender, axisTitleMark, this);
    }
    renderDom(axisTitleMark, axisWrap);
    let _labelNode = axisWrap.selectAll('text.axislabel'),
        _label = _labelNode.node();

    axes.axisLabel = _label;
    // if (axes.axisLabel && axes.axisLabel.order == order) {
    //     let axisAreaDim = getAxisAreaDim(type, order, orient, chart, plotRect);
    //     axes.axisLabel.show = !hideTitle;
    //     axes.axisLabel.transformLabel(axisgroup, axisAreaDim, orient);
    // } else {
    //     axes.axisLabel = new axis.AxisLabel(axisdata_merged, axisLabelProperties, orient, axisWrap);
    // }

    // renderDom(marks, tempDom);

    if (!hideTitle) {
        if (!_labelNode.empty() && axisWrap.style('display') !== NONE) {
            axisSize = _label.getBBox().height; // calculate and cache.
        } else {
            axisSize = 0;
        }
        // axisSize = _label.getBBox().height;
        axislabelProp.textAnchor = 'middle';
        switch (orient) {
            case 'left': {
                var width_new = tickPrecalculation.inlineToPlot
                    ? 0
                    : tickPrecalculation.isInline
                    ? axisgroup.width
                    : axisgroup.width + tickMargin.left;
                preCalcAxisLen = width_new + titleMargin.right + axisSize + titleMargin.left;
                break;
            }
            case 'right': {
                var width_new = tickPrecalculation.inlineToPlot
                    ? 0
                    : tickPrecalculation.isInline
                    ? axisgroup.width
                    : axisgroup.width + tickMargin.right;
                preCalcAxisLen = width_new + titleMargin.left + axisSize + titleMargin.right;
                break;
            }
            case 'bottom': {
                var height_new = tickPrecalculation.inlineToPlot
                    ? 0
                    : tickPrecalculation.isInline
                    ? axisgroup.height
                    : axisgroup.height + tickMargin.bottom;
                preCalcAxisLen = height_new + titleMargin.bottom + axisSize + titleMargin.top;
                break;
            }
            case 'top': {
                var height_new = tickPrecalculation.inlineToPlot
                    ? 0
                    : tickPrecalculation.isInline
                    ? axisgroup.height
                    : axisgroup.height + tickMargin.top;
                preCalcAxisLen = height_new + titleMargin.top + axisSize + titleMargin.bottom;
                break;
            }
        }
    } else {
        preCalcAxisLen = orient === 'bottom' || orient == 'top' ? axisgroup.height + tickMargin.bottom : axisgroup.width + tickMargin.left;
    }

    // let width = axisgroup.width,
    //     height = axisgroup.height;
    // if ((orient === 'left' || orient === 'right') && defined(tick_data.width)) {
    //     width = perToPx(getPriorityValue1(tick_data.width, tick_data.maxWidth) || axis.ticks.maxTickSizeRatio_Hori, plotRect.width);
    // } else if (defined(this.axisTicks.tickInfo.height)) {
    //     height = perToPx(getPriorityValue1(tick_data.height, tick_data.maxHeight) || axis.ticks.maxTickSizeRatio_Vert, plotRect.height);
    // }
    // _label.remove();

    // let preCalcTickLen = {
    //     x: axisgroup.left,
    //     y: axisgroup.top,
    //     width: width,
    //     height: height
    // };
    return {
        // preCalcTickLen,
        preCalcAxisLen
    };
    // return preCalcAxisLen;
}
