// @zohocharts/zc/src/axis/ticks/custom-ticks.js
import { defined, invokeFunction, isArray, pick, toArray } from '../../utils/zcutils';
import { createElement, parseSVGElementsToArray, SVGDOMParser } from '../../utils/zcsvgutils';

export function customTicks(ticks) {
    let { type, order, chart, config, plotRect, datatype, column } = this;

    let tickConfig = config.ticklabel;
    var parentArray = [],
        parentStyles = [],
        userTickValues = [],
        hasCustomG;

    var userColumnInfo = {
        data: [column]
    };
    column.columnindex = pick(column.columnindex, 0); //TODO pass proper columindex instead of 0

    ticks.forEach(function(d, i) {
        var userInput = invokeFunction(tickConfig.showAs, d, tickConfig, userColumnInfo, chart, i, order),
            childArray = [],
            styles = [];

        if (isArray(userInput)) {
            parentArray.push(childArray);
            parentStyles.push(styles);
        } else {
            childArray = parentArray;
            styles = parentStyles;
        }
        userTickValues.push(userInput);

        if (defined(userInput)) {
            // if (isArray(userInput)) {
            //     userInput = userInput.join('\n');
            // }
            // var sandBox = document.createElement('div');
            // sandBox.innerHTML = userInput;

            // let userInputWrapped = `<zc-temp>${userInput}</zc-temp>`,
            // sandBox = new DOMParser().parseFromString(userInputWrapped, 'text/xml');
            let sandBox = SVGDOMParser(userInput);

            var children = sandBox.querySelectorAll('zc-temp')[0].childNodes,
                tspan = parseSVGElementsToArray(sandBox, 'tspan'),
                images = parseSVGElementsToArray(sandBox, 'image,img'),
                group = parseSVGElementsToArray(sandBox, 'g');

            if (children && children.length) {
                for (let i = 0; i < children.length; i++) {
                    if (i > 0 && !tspan.length && !images.length) {
                        return;
                    }
                    let t = children[i];
                    if (t.tagName == 'tspan') {
                        childArray.push(t.textContent.trim());
                        let style = t.style || {},
                            attributes = t.attributes || {};
                        styles.push({
                            fontSize: style.fontSize && parseFloat(style.fontSize),
                            fontFamily: style.fontFamily,
                            fontStyle: style.fontStyle,
                            fontWeight: style.fontWeight,
                            x: attributes.x,
                            y: attributes.y,
                            dy: attributes.dy
                        });
                    } else if (t.tagName == 'image' || t.tagName == 'img') {
                        toArray(sandBox.querySelectorAll('image,img')).forEach(function(t) {
                            let style = t.style || {},
                                attributes = t.attributes || {};
                            childArray.push({
                                width: +t.getAttribute('width'),
                                height: +t.getAttribute('height')
                            });
                            styles.push({
                                y: t.getAttribute('y'),
                                dy: t.getAttribute('dy')
                            });
                        });
                    } else if (t.tagName == 'g') {
                        hasCustomG = true;
                        let customG = createElement('g', '.customTick', [1], chart.wrap, null).html(userInput),
                            bbox = customG.node().getBBox();

                        childArray.push({
                            width: +bbox.width,
                            height: +bbox.height,
                            type: 'group'
                        });
                        customG.remove();
                    } else if (!tspan.length && !images.length && !group.length) {
                        childArray.push(userInput);
                    }
                }
            } else {
                childArray.push(userInput);
            }

            // if (tspan.length) {
            //     tspan.forEach(function(t) {
            //         childArray.push(t.textContent.trim());
            //         let style = t.style || {},
            //             attributes = t.attributes || {};
            //         styles.push({
            //             fontSize: style.fontSize && parseFloat(style.fontSize),
            //             fontFamily: style.fontFamily,
            //             fontStyle: style.fontStyle,
            //             fontWeight: style.fontWeight,
            //             x: attributes.x,
            //             y: attributes.y,
            //             dy: attributes.dy
            //         });
            //     });
            // } else if (!images.length) {
            //     childArray.push(userInput);
            // }
            //
            // if (images.length) {
            //     toArray(sandBox.querySelectorAll('image,img')).forEach(function(t) {
            //         let style = t.style || {},
            //             attributes = t.attributes || {};
            //         childArray.push({
            //             width: +t.getAttribute('width'),
            //             height: +t.getAttribute('height')
            //         });
            //         styles.push({
            //             y: t.getAttribute('y'),
            //             dy: t.getAttribute('dy')
            //         });
            //     });
            // }
            sandBox = null;
        }
    });

    return {
        customGroup: hasCustomG,
        tickValues: parentArray,
        styles: parentStyles,
        userRawTicks: userTickValues
    };
}
