// @zohocharts/zc/src/axis/ticks/log-ticks.js
import { d3_extent, d3_min } from '../../core/d3-constants';
import { pick, arrayIncludes, defined } from '../../utils/zcutils';
// import { getMargin } from '../axisutils';
import { getLogTicks } from '../log-tick-labels';
import { addMinmaxToTicks } from "./min-max-ticks";
import { NULL, UNDEFINED } from '../../core/zc-constants';
import { LOG } from '../../core/scale-type-constant';
import { _hasDefaultBase } from '../../core/data-processor/minMaxYValues';
import { getAxisObject } from '../axisutils';
import { scale } from '../../core/main-zc-init';


export const logTicks = scale.log.ticks;
export const _getLogTicks = logTicks._getLogTicks = function _getLogTicks(logTickH) {
    {
        let { scale, size, side, styles, format, rotation, formatter, formatAccessor, margin } = this,
            { domain } = logTickH,
            { tickCount, tickInterval } = this.config.ticklabel,
            { column, orient } = this,
            tickValues,
            minorTickValues,
            logDomain,
            range,
            formattingSubFunction = format.subfunction;

        if (scale.constant) {
            let min = d3_min(domain);
            if (min > 0 && min < pick(logTickH.axisObj.logBase, 10) && logTickH.axisObj.allowNonPositiveTicks === false) {
                scale.constant(0.0001);
            } else {
                scale.constant(1);
            }
        }
        const addMinMaxLog = tickValues => {
            logTickH.axisInfo.logDomain = d3_extent(tickValues);

            return addMinmaxToTicks.call(this, tickValues);
        };
        const formatFunc = tickValues => {
            return tickValues.map(formatAccessor);
        };
        // fix all negative
        ({ tickValues, minorTickValues, logDomain } = getLogTicks(
            logTickH.domainCopy,
            size,
            logTickH.dataMinMax,
            logTickH.usePadding,
            logTickH.axisObj,
            scale.copy(),
            margin,
            {
                fontSize: styles.fontSize,
                fontFamily: styles.fontFamily
            },
            formatFunc,
            orient,
            rotation,
            logTickH.adjustToBase,
            addMinMaxLog,
            logTickH.hasFixedPosition,
            domain
        ));
        logTickH.axisInfo.logDomain = logDomain;
        return {ticks: tickValues, minorTickValues};

    }
}

export const _logDomain = scale.log.domain = function _logDomain(helpers) {
    let { type, order, chart, column, config } = helpers,
        { dataObject } = chart,
        axisInfo = getAxisObject(chart, type, order);
    // clear the existing domain
    axisInfo.logDomain = null;
        let extent = type === 'x' && order === 0 ? chart.viewPortExtent : NULL,
        domain = defined(extent)
            ? extent
            : type === 'x'
            ? dataObject.getMinMaxX()
            : dataObject.getMinMaxY(order, UNDEFINED, chart.datalabels.hasFixedPosition);



    // #ZC2805 $domain changed to $dataMinMax
    let dataMinMax = type === 'x' ? dataObject.getMinMaxX(0) : dataObject.getMinMaxY(order, 0),
        addBaseline = !arrayIncludes(dataMinMax),
        domainCopy = dataMinMax.slice(0),
        logDomain,
        usePadding = !(dataObject.isPercentChart || dataObject.hasLevelMarker || dataObject.isWaterfallEnabled),
        adjustToBase = null,
        yEncode = dataObject.getEncode('y', order);

    // FEATURES_LOGAXIS_45
    if (dataMinMax[0] === 0 && dataMinMax[1] === 0) {
        column.scaleType = null;
        domain = type === 'x' ? dataObject.getMinMaxX(0) : dataObject.getMinMaxY(order, UNDEFINED, chart.datalabels.hasFixedPosition);
        column.scaleType = LOG;
    }

    // type y case CXZA_LEGEND_26
    if (yEncode && _hasDefaultBase(yEncode, chart.dataObject, chart) && addBaseline && type === 'y' && !defined(config.minRange)) {
        adjustToBase = 0;
        domainCopy.push(0);
        domainCopy = d3_extent(domainCopy);
    }

    let logTickH = {
        dataMinMax: dataMinMax,
        usePadding: usePadding,
        axisObj: config,
        adjustToBase: adjustToBase,
        hasFixedPosition: chart.datalabels.hasFixedPosition,
        axisInfo,
        domainCopy,
        domain
    };
    dataObject.logTickH = logTickH;
    return logTickH;
}
