// @zohocharts/zc/src/axis/ticks/longest-tick.js
import { defined, perToPx, sortByIndices } from '../../utils/zcutils';
import { getTextBound } from '../textBound';
import { renderDom } from '../../mark/dom-render';
import { getTextContent, wrapTextContent } from '../../utils/zcsvgutils';
import { tick_overflow_helpers } from './tick-overflow-helpers';
import { handleSVGTextOverflow } from './tick-overflow';
import { axis } from '../../core/main-zc-init';
import { MarksRegistry } from '../../plot/MarksRegistry';

export function precalculateTickGroup(marks, tempDom) {
    let { plotRect, config, orient, datatype, chart } = this,
        { scale, styles, chartArea, plotSize } = this.axisTicks,
        { isHierarchialLevelType, isHierarchy } = this.axisTicks.tickH,
        { rotation, ticks, ticks_formatted, customTickInfo, longestTick } = this.axisTicks.tickInfo,
        useHTML = customTickInfo && customTickInfo.styles,
        // tempPlaceHolder,
        left = orient == 'right' ? plotRect.left + plotRect.width : plotRect.left,
        top = orient == 'bottom' ? plotRect.top + plotRect.height : plotRect.top;

    let ticks_sorted_index = sortByIndices(ticks_formatted),
        counter = -1,
        isFinalLongTick,
        curIdx = ticks_sorted_index[0],
        maxIdx = ticks_sorted_index[0],
        sortIdx,
        curTick,
        curTick_formatted,
        curTick_content,
        maxTick,
        maxTick_formatted,
        maxTick_content,
        maxTick_htmlContent,
        nextTick,
        textWidth = textWidthAccessor({ orient, customTickInfo, styles });

    var markSelection = MarksRegistry.get('markSelection');
    var markSelector = new markSelection(marks);

    if (useHTML) {
        //Need to sort the ticklabels by finding text width
        var formattedCustomTickValuesSize = ticks_formatted.map((d, i) => {
            return textWidth(d, i);
        });
        ticks_sorted_index = sortByIndices(formattedCustomTickValuesSize, 'integer');
    }
    let _helperObj = { config, orient, isHierarchialLevelType, isHierarchy, tempDom, rotation, ticks, styles, chart, scale, plotSize },
        overflowH = tick_overflow_helpers.call(this.axisTicks, { isHierarchialLevelType, isHierarchy, tempDom });

    /*
    Initiate max tick content

     */
    maxIdx = longestTick.index;
    maxTick = ticks[maxIdx];
    maxTick_formatted = useHTML ? customTickInfo.userRawTicks[longestTick.index] : longestTick.formattedVal;

    markSelector.select('selector', '.zc-tick').data([maxTick]);
    markSelector.select('name', 'ticklabel').data([maxTick_formatted]);
    marks = markSelector.parser;
    // marks.children[0].data = [maxTick];
    // marks.children[0].children[0].data = [maxTick_formatted];
    renderDom(marks, tempDom);
    wrapTextContent(tempDom.selectAll('text'), maxTick_formatted, null, useHTML); //TODO change for custom tick and merge into renderDOM
    handleSVGTextOverflow(_helperObj, overflowH, { node: tempDom.selectAll('text')._groups[0][1] || tempDom.selectAll('text').node(), index: maxIdx });
    maxTick_content = getTextContent(tempDom.selectAll('text'));
    maxTick_htmlContent = useHTML ? getCustomTickContent(tempDom) : maxTick_content;

    while (!isFinalLongTick && counter < ticks.length - 2) {
        //TODO in previous we didnt consider for lwast sorted tick
        counter = counter + 1;

        sortIdx = ticks_sorted_index[counter];
        curTick = ticks[sortIdx];
        curTick_formatted = useHTML ? customTickInfo.userRawTicks[sortIdx] : ticks_formatted[sortIdx];

        // maxTick = ticks[maxIdx];
        // maxTick_formatted = useHTML ? customTickInfo.userRawTicks[maxIdx] : ticks_formatted[maxIdx];
        // let tickContent = useHTML ? : curTick;
        //apply marks
        markSelector.select('selector', '.zc-tick').data([curTick]);
        markSelector.select('name', 'ticklabel').data([curTick_formatted]);
        marks = markSelector.parser;

        // marks.children[0].data = [curTick];
        // marks.children[0].children[0].data = [curTick_formatted];
        renderDom(marks, tempDom);
        wrapTextContent(tempDom.selectAll('text'), curTick_formatted, null, useHTML); //TODO change for custom tick and merge into renderDOM
        handleSVGTextOverflow(_helperObj, overflowH, { node: tempDom.selectAll('text')._groups[0][1] || tempDom.selectAll('text').node(), index: counter });
        curTick_content = getTextContent(tempDom.selectAll('text'));
        // if (counter == 0) {
        //     maxTick_content = curTick_content;
        // }else{
        if (textWidth(maxTick_content, maxIdx) > textWidth(curTick_formatted, sortIdx)) {
            isFinalLongTick = true;
        }
        // }
        // handleOverflowTicks();
        // if(textWidth(maxTick_content, maxIdx) < textWidth(curTick_formatted, sortIdx)){
        //     isFinalLongTick = true;
        // }
        if (textWidth(curTick_content, curIdx) > textWidth(maxTick_content, maxIdx)) {
            maxTick = ticks[sortIdx];
            maxTick_formatted = useHTML ? customTickInfo.userRawTicks[sortIdx] : ticks_formatted[sortIdx];
            maxIdx = sortIdx;
            maxTick_content = curTick_content;
            maxTick_htmlContent = useHTML ? getCustomTickContent(tempDom) : maxTick_content;
        } else {
            // isFinalLongTick = true;
        }
        //check overflow and next tick
        var isLastTick = !ticks.length || counter === ticks.length - 1 || ticks.length === 1;
        tempDom.selectAll('g.customTick').remove();
    }

    markSelector.select('selector', '.zc-tick').data([maxTick]);
    markSelector.select('name', 'ticklabel').data([maxTick_formatted]);
    marks = markSelector.parser;

    // marks.children[0].data = [maxTick];
    // marks.children[0].children[0].data = [maxTick_formatted];
    renderDom(marks, tempDom);

    //2.24 viewparser
    
    // maxTick_htmlContent = useHTML ? getCustomTickContent(tempDom) : tempDom.selectAll('text.ticklabel').text();

    wrapTextContent(tempDom.selectAll('text'), maxTick_htmlContent, null, useHTML); //TODO change for custom tick and merge into renderDOM
    //calculate g.customTick for custom group tick element
    let finalWrap =
        useHTML && customTickInfo.tickValues[maxIdx] && customTickInfo.tickValues[maxIdx].type == 'group'
            ? tempDom.selectAll('g.customTick')
            : tempDom.selectAll('text');

    let axisgroup = tempDom.node().getBoundingClientRect();
    let width = axisgroup.width,
        height = axisgroup.height,
        tick_data = this.axisTicks.tickInfo;
    if ((orient === 'left' || orient === 'right') && defined(tick_data.width)) {
        width = perToPx(getPriorityValue1(tick_data.width, tick_data.maxWidth) || axis.ticks.maxTickSizeRatio_Hori, plotRect.width);
    } else if (defined(this.axisTicks.tickInfo.height)) {
        height = perToPx(getPriorityValue1(tick_data.height, tick_data.maxHeight) || axis.ticks.maxTickSizeRatio_Vert, plotRect.height);
    }

    let preCalcTickLen = {
        x: axisgroup.left,
        y: axisgroup.top,
        width: width,
        height: height
    };

    return {
        axisgroup,
        preCalcTickLen,
        longTickSize: finalWrap.node().getBoundingClientRect(),
        ticks
    };
}

var getPriorityValue1 = function(constant, max) {
    if (defined(constant) && defined(constant)) {
        return constant > max ? max : constant;
    } else {
        return constant || max;
    }
};

function textWidthAccessor(tickInfo) {
    let { customTickInfo, styles, orient } = tickInfo,
        useHTML = customTickInfo && customTickInfo.styles;
    var style, dimension;
    return function(text, index) {
        if (useHTML) {
            style = customTickInfo.styles[index];
        }

        dimension = getTextBound(text, styles.fontSize, styles.fontFamily, style);

        if (useHTML) {
            return dimension.width * dimension.height;
        }

        return orient == 'left' || orient == 'right' ? dimension.width : dimension.width * dimension.height;
    };
}
function getCustomTickContent(actualTickGroup) {
    var tspan = actualTickGroup
            .selectAll('text.ticklabel')
            .selectAll('tspan')
            .nodes(),
        image = actualTickGroup
            .selectAll('g.image')
            .selectAll('image', 'img')
            .nodes(),
        group = actualTickGroup.selectAll('g.customTick').nodes(),
        childArray = [];

    if (group && group.length) {
        var arr = [];
        group.forEach(function(d) {
            childArray.push(d.outerHTML);
        });
    } else {
        if (tspan && tspan.length) {
            var arr = [];
            tspan.forEach(function(d) {
                childArray.push(d.outerHTML);
            });
        }
        if (image && image.length) {
            var arr = [];
            image.forEach(function(d) {
                childArray.push(d.outerHTML);
            });
        }
    }

    return childArray;
}
