// @zohocharts/zc/src/axis/ticks/numeric-ticks.js
import { mathCeil, mathFloor, mathLN10, mathLog, mathMax, mathPow, mathRound } from '../../core/zc-constants';
import { d3_range, d3_tickStep } from '../../core/d3-constants';
import { defined, invokeFunction, pick } from '../../utils/zcutils';
import { nearestTickInterval, nearestNiceMin } from '../tick-labels';
import { addMinmaxToTicks } from './min-max-ticks';

export function getNumericTicks() {
    let { domain, simpleScale, size, side, format } = this,
        { tickCount, interval, maxRange, tickMultiples, minTick } = this.config,
        ticks,
        range,
        formattingSubFunction = format.subfunction,
        calculatedInterval = (domain[1] - domain[0]) / interval;

    if (simpleScale.isD3) {
        simpleScale = simpleScale()
            .domain(domain)
            .range(size);
    }
    if (defined(interval) && calculatedInterval < size) {
        interval = defined(tickCount) ? (domain[1] - domain[0]) / tickCount : interval;
        ticks = getLinearTicksByInterval(domain, interval, minTick, maxRange);
    } else if (+tickCount > 0) {
        interval = (domain[1] - domain[0]) / tickCount;
        ticks = getLinearTicksByCount(domain, tickCount, minTick, maxRange, formattingSubFunction, tickMultiples);
    } else if (formattingSubFunction === 'integer') {
        ticks = getLinearTicksBySubFunction(domain, null, maxRange);
    } else {
        if (side === 'height') {
            tickCount = tickCount || mathRound(size / 40);
            range = [size, 0];
        } else {
            tickCount = tickCount || mathRound(size / 100);
            range = [0, size];
        }
        if (defined(minTick)) {
            let step = d3_tickStep(minTick, domain[1], tickCount);
            ticks = d3_range(minTick, domain[1], step);
        } else {
            ticks = simpleScale.ticks(tickCount);
        }
    }
    ticks = addMinmaxToTicks.call(this, ticks);


    return ticks;
}

export function getLinearTicksByCount(domain, count, customStart, customStop, intervalType, tickMultiples) {
    if (count === 1) {
        step = getLinearStep(domain, 1);
        min = Math.ceil(domain[0] / step) * step;
        return [min];
    } else {
        var tickInterval = (domain[1] - domain[0]) / (count - 1),
            min,
            max,
            remaining,
            m = count - 1,
            span = domain[1] - domain[0],
            step = span / m,
            power = mathFloor(mathLog(step) / mathLN10),
            magnitude = mathPow(10, power),
            ticks;

        magnitude = magnitude * Math.floor(step / magnitude);
        magnitude = +magnitude.toFixed(12); //To avoid floating point error (e.g):1.7000000000000002

        var nicedInterval = nearestTickInterval(tickInterval, magnitude, tickMultiples, intervalType);

        if (intervalType == 'integer' && nicedInterval < 1) {
            return getLinearTicksBySubFunction(domain, count, customStop);
        }
        min = domain[0];
        max = domain[1];
        remaining = span - nicedInterval * m;

        // if (remaining < nicedInterval && Math.floor((span / m) / magnitude) > 1) {
        //     nicedInterval -= magnitude;
        //     remaining = span - (nicedInterval * m);
        // }

        var mag_remaining = mathPow(10, mathFloor(mathLog(remaining) / mathLN10) + 1);
        min = nearestNiceMin(min, remaining, mag_remaining, nicedInterval);
        min = +min.toFixed(12); //To avoid floating point error (e.g):1.7000000000000002
        min = pick(customStart, min);
        max = min + nicedInterval * count;
        max = +max.toFixed(12); //Test_559,560

        max = max > domain[1] ? +domain[1] : max;
        ticks = d3_range(min, max, nicedInterval);
        if (defined(customStop)) {
            ticks = addMaxStopToTicks(ticks, nicedInterval, customStop);
        } else if (ticks.length < count) {
            ticks = addMaxStopToTicks(ticks, nicedInterval, max);
        }
        return ticks;
    }
}

function getLinearStep(domain, count) {
    var m = defined(count) ? count : 10,
        span = domain[1] - domain[0],
        step = mathPow(10, mathFloor(mathLog(span / m) / mathLN10)),
        err = (m / span) * step;
    if (err <= 0.15) {
        step *= 10;
    } else if (err <= 0.35) {
        step *= 5;
    } else if (err <= 0.75) {
        step *= 2;
    }
    return step;
}

export function getLinearTicksByInterval(domain, tickInterval, customStart, customStop) {
    let count = mathFloor((domain[1] - domain[0]) / tickInterval),
        min,
        ticks;

    if (count) {
        let step = getLinearStep(domain, count);
        min = mathCeil(domain[0] / step) * step;

        if (min - domain[0] > tickInterval) {
            var s = mathPow(10, mathFloor(mathLog(tickInterval) / mathLN10)),
                q = mathFloor((min - domain[0]) / s);
            min -= q * s;
        }
    } else {
        var span = domain[1] - domain[0],
            s = mathPow(10, mathFloor(mathLog(span) / mathLN10));
        min = domain[0] - (domain[0] % s) + s;
    }
    min = pick(customStart, min);
    ticks = d3_range(min, domain[1], tickInterval);
    if (defined(customStop)) {
        ticks = addMaxStopToTicks(ticks, tickInterval, customStop);
    }
    return ticks;
}

export function getLinearTicksBySubFunction(domain, count, customStop) {
    let ticks,
        interval = getNiceStep(domain, count);
    ticks = d3_range(getNiceMin(domain, 1), domain[1], interval);
    if (defined(customStop)) {
        ticks = addMaxStopToTicks(ticks, interval, customStop);
    }
    return ticks;
}

function getNiceMin(domain, minStep) {
    var step = getLinearStep(domain);
    if (defined(minStep)) {
        // Test_525.json
        step = mathMax(minStep, step);
    }
    return mathCeil(domain[0] / step) * step;
}

function getNiceStep(domain, count) {
    return mathCeil(getLinearStep(domain, count));
}

function addMaxStopToTicks(ticks, interval, stopValue) {
    if (ticks[ticks.length - 1] + interval <= stopValue) {
        ticks.push(ticks[ticks.length - 1] + interval);
    }
    return ticks;
}
