// @zohocharts/zc/src/axis/ticks/tick-overflow.js
import { d3_select } from '../../core/d3-constants';
import { mathAbs, mathCeil, mathMax, mathMin } from '../../core/zc-constants';
import { getDefaultBoundingClientRect, getTspanBoundingClientRect } from '../textBound';
import { naiveEllipsis, trimText } from '../../utils/zcsvgutils';
import { getBoundingRect } from '../../utils/zcutils';

export function handleSVGTextOverflow(helpers, overflowH, nodeH) {
    let { plotRect, side, overflow, orient, rotation, scale } = helpers,
        maxTickSize_Hori = overflowH.maxTickSize.horizontal,
        maxTickSize_Vert = overflowH.maxTickSize.vertical,
        maxTickSize,
        { margin, boundaryEdges, prerender, needTrimByMaxSize, tickSpacing, hasDynamicTickSpacing, minMargin, overflowPattern } = overflowH,
        { node, index } = nodeH;
    if (overflow == 'none') {
        return;
    }
    /*
     * Cropping the text to maximum ticklabel size (according to bounded box of the text)
     */
    var activeNode = node;
    var childNodes = [activeNode.childNodes[0]]; //If the SVG text has title element, the title element will be removed while changing this.textContent as because it modifies all the childNodes of it. So take the text element alone to trim

    // only for multiple line ticklabel, checking tspan
    var tspan = d3_select(activeNode)
        .selectAll('tspan')
        .nodes();
    if (tspan && tspan.length) {
        tspan.sort(function(a, b) {
            return b.textContent.length - a.textContent.length;
        });
        childNodes = tspan;
    }

    if (!childNodes) {
        return;
    }

    //Change maxTickSize if margin was not needed
    if (overflowH.fitTextInBetween) {
        //dontUseMargin
        maxTickSize = overflowH.dynamicMaxTickSize()(node, index, helpers, overflowH);
        // let count = node.selectAll('text.ticklabel').nodes().length;
        // var outerPadding = mathAbs(distBwTicks.outerPaddingInPixel);
        // if (orient == 'bottom' || orient == 'top') {
        //     if (hasUserLabelSize) {
        //         maxTickSize_Hori = labelSize.horizontal;
        //     } else if (rotation != 90 && rotation != 270) {
        //         maxTickSize_Hori = index == 0 || index == count - 1 ? outerPadding * 2 : mathAbs(distBwTicks.rangeBand) * 2;
        //     }
        // } else if ((orient == 'left' || orient == 'right') && (rotation == 90 || rotation == 270)) {
        //     maxTickSize_Vert = index == 0 || index == count - 1 ? outerPadding * 2 : mathAbs(distBwTicks.rangeBand) * 2;
        // }
    }

    //Iterate childnodes and apply ellipsis for each node
    for (var i = 0; i < childNodes.length; i++) {
        var txtNode = tspan && tspan.length ? childNodes[i] : activeNode,
            initial_textContent = txtNode.textContent,
            clientRect = txtNode.nodeName == 'tspan' ? getTspanBoundingClientRect : getDefaultBoundingClientRect,
            textBounds = clientRect(txtNode),
            initial_textBound = textBounds,
            trimmed,
            overflowLength = -1,
            max = {
                // This should be changed for axis labels
                width: maxTickSize_Hori,
                height: maxTickSize_Vert
            };

        //For hierarchy level type, check only for maxWidth
        if (hasDynamicTickSpacing && needTrimByMaxSize && !prerender) {
            var rb =
                orient == 'left' || orient == 'right' ? maxTickSize_Hori : Math.abs(scale.discreteWidth(node.__data__)) || maxTickSize_Hori;
            var maxWid = mathMin(maxTickSize_Hori, rb);
            naiveEllipsis(txtNode, txtNode, txtNode.textContent, maxWid, maxTickSize_Vert, textBounds, overflowPattern);
            return;
        }

        //Check for maxwidth and add ellipsis
        if (needTrimByMaxSize) {
            //labelType == 'ticklabel'
            let trimmedNode = trimTextBasedOnMaxSize(
                txtNode,
                textBounds,
                { horizontal: maxTickSize_Hori, vertical: maxTickSize_Vert },
                overflowPattern
            );
            textBounds = clientRect(trimmedNode);
            // if (maxTickSize_Hori < textBounds.width) {
            //     overflowLength = (textBounds.width - maxTickSize_Hori) / textBounds.width;
            //     var croppedTextLength = mathCeil((1 - overflowLength) * txtNode.textContent.length);
            //     txtNode.textContent = txtNode.textContent.substring(0, croppedTextLength - 2) + overflowPattern; //TODO: for textOverflow = 'hidden', subtracting 2 letters is necessary?
            //     textBounds = clientRect(txtNode); //update new textbounds
            // }
            // if (maxTickSize_Vert < textBounds.height) {
            //     overflowLength = (textBounds.height - maxTickSize_Vert) / textBounds.height;
            //     var croppedTextLength = mathCeil((1 - overflowLength) * txtNode.textContent.length);
            //     txtNode.textContent = txtNode.textContent.substring(0, croppedTextLength - 2) + overflowPattern;
            //     textBounds = clientRect(txtNode); //update new textbounds
            // }
        }

        //calculte new boundary edges and textbounds

        // edge = overflowH.boundaryEdges();
        var referenceBounds = overflowH.referenceBounds();
        var newTextBounds = {
            left: textBounds.left - referenceBounds.left,
            right: textBounds.left + textBounds.width - referenceBounds.left,
            top: textBounds.top - referenceBounds.top,
            bottom: textBounds.top + textBounds.height - referenceBounds.top,
            width: textBounds.width,
            height: textBounds.height
        };

        // if (options.edge) {
        //     boundaryEdges = options.edge(txtNode, textBounds);
        // } else
        if (overflowH.needDynamicEdge && scale.bandwidth) {
            boundaryEdges = overflowH.dynamicBoundaryEdges()(node, index, helpers, overflowH, textBounds);
        }

        //Trim text based on boundary edges
        if ((orient == 'left' || orient == 'right') && (rotation != 0 && rotation != 180 && rotation != 360)) {
            if (
                newTextBounds.top + mathMax(minMargin, margin.top) < boundaryEdges.top ||
                newTextBounds.bottom - mathMax(minMargin, margin.bottom) > boundaryEdges.bottom
            ) {
                trimmed = true;
                trimText(txtNode, txtNode.textContent, newTextBounds, boundaryEdges, orient, overflowPattern, margin, max, txtNode);
            }
        } else if ((orient == 'bottom' || orient == 'top') && (rotation != 90 && rotation != 270)) {
            if (
                newTextBounds.left + mathMax(minMargin, margin.left) < boundaryEdges.left ||
                newTextBounds.right - mathMax(minMargin, margin.right) > boundaryEdges.right
            ) {
                trimmed = true;
                trimText(txtNode, txtNode.textContent, newTextBounds, boundaryEdges, orient, overflowPattern, margin, max, txtNode);
            }
        }

        if (needTrimByMaxSize && !trimmed) {
            var textBounds = clientRect(txtNode);
            if (textBounds.width != initial_textBound.width || textBounds.height != initial_textBound.height) {
                naiveEllipsis(txtNode, activeNode, initial_textContent, max.width, max.height, textBounds, overflowPattern);
            }
        }

        if (txtNode.textContent == overflowPattern) {
            txtNode.textContent = prerender ? '.' : ''; //hack  while casper automation DOC_CHART_61
        }
    }
}

function trimTextBasedOnMaxSize(node, bounds, maxSize, overflowPattern) {
    let clientRect = node.nodeName == 'tspan' ? getTspanBoundingClientRect : getDefaultBoundingClientRect;
    if (maxSize.horizontal < bounds.width) {
        let overflowLength = (bounds.width - maxSize.horizontal) / bounds.width;
        var croppedTextLength = mathCeil((1 - overflowLength) * node.textContent.length);
        node.textContent = node.textContent.substring(0, croppedTextLength - 2) + overflowPattern; //TODO: for textOverflow = 'hidden', subtracting 2 letters is necessary?
        bounds = clientRect(node);
    }
    if (maxSize.vertical < bounds.height) {
        let overflowLength = (bounds.height - maxSize.vertical) / bounds.height;
        var croppedTextLength = mathCeil((1 - overflowLength) * node.textContent.length);
        node.textContent = node.textContent.substring(0, croppedTextLength - 2) + overflowPattern;
    }
    return node;
}

function trimTextBasedOnEdges() {}
