// @zohocharts/zc/src/axis/ticks/tick-overflow-helpers.js
import { mathAbs } from '../../core/zc-constants';
import { defined, getBoundingRect, pick } from '../../utils/zcutils';
import { scale } from '../../core/main-zc-init';
import { getMaxTickSize } from '../tick-helpers';
import { getPlotSize } from '../../plot/plotSize';

let axislabel_overflow_helpers = helpers => {
    let { scale, ticks, rotation, config, chartRect, orient, type } = helpers;
    let overflow_helpers = {
        needTrimByMaxSize: false,
        maxTickSize: getMaxTickSize(config.ticklabel, chartRect),
        boundaryEdges: getBoundaryEdgesForTick(),
        hasDynamicTickSpacing: helpers.isHierarchialLevelType,
        fitTextInBetween: !helpers.isHierarchialLevelType && ((helpers.isHierarchy() && type == 'x') || config.tickmark == 'between'),
        tickSpacing: distanceBetweenTicks(),
        margin: getExtraMarginForBoundary(helpers),
        dynamicBoundaryEdges: () => getDynamicBoundaryEdgesForTick(),
        dynamicMaxTickSize: () => getMaxTickSizeByData(),
        referenceBounds: () => getBoundingRect(helpers.tempDom.selectAll('.domain'))
    };

    return overflow_helpers;
};

export function tick_overflow_helpers(_h) {
    let { scale, config, chartRect, orient, type, chart, styles, plotSize } = this,
        { rotation, ticks, ticks_formatted, customTickInfo, longestTick } = this.tickInfo,
        helpers = { ..._h, scale, ticks, rotation, config, chartRect, orient, type, chart, styles, plotSize};
    let overflow_helpers = {
        needTrimByMaxSize: true,
        hasDynamicTickSpacing: helpers.isHierarchialLevelType(),
        fitTextInBetween: fitTextInBetween(helpers),
        maxTickSize: getMaxTickSize(config.ticklabel, chartRect),
        boundaryEdges: getBoundaryEdgesForTick(helpers),
        tickSpacing: distanceBetweenTicks(helpers),
        minMargin: needMinMarginToExcludeFromTrim(helpers) ? helpers.styles.fontSize : 0,
        margin: getExtraMarginForBoundary(helpers),
        needDynamicEdge: chart && chart.dataObject && chart.dataObject.hierarchical && chart.dataObject.hierarchical.dynamicLeafScale,
        dynamicBoundaryEdges: () => getDynamicBoundaryEdgesForTick,
        dynamicMaxTickSize: () => getMaxTickSizeByData,
        referenceBounds: () => getBoundingRect(helpers.tempDom.selectAll('.axisline')),
        overflowPattern: config.ticklabel.textOverflow == null || config.ticklabel.textOverflow == 'ellipsis' ? '..' : '',
        prerender: true
    };

    return overflow_helpers;
};
function needMinMarginToExcludeFromTrim(helpers) {
    if (fitTextInBetween(helpers) || helpers.chart.dataObject.isPolarAxisCategory) {
        return false;
    } else {
        return true;
    }
}
function fitTextInBetween(helpers) {
    //dontUseMargin
    return (!helpers.isHierarchialLevelType() && (helpers.isHierarchy() && helpers.type == 'x')) || helpers.config.tickmark == 'between';
}
function distanceBetweenTicks(helpers) {
    if (fitTextInBetween(helpers)) {
        return scale.getRangeBand(helpers.scale, helpers.ticks, 0);
    }
}
function getMaxTickSizeByData(node, index, helpers, overflowH) {
    let maxTickSize_Hori, maxTickSize_Vert;
    let count = helpers.tempDom.selectAll('text.ticklabel').nodes().length;
    var outerPadding = mathAbs(overflowH.tickSpacing.outerPaddingInPixel),
        hasUserLabelSize = defined(helpers.config.ticklabel.width) || defined(helpers.config.ticklabel.maxWidth);
    if (helpers.orient == 'bottom' || helpers.orient == 'top') {
        if (hasUserLabelSize) {
            maxTickSize_Hori = overflowH.maxTickSize.horizontal;
        } else if (helpers.rotation != 90 && helpers.rotation != 270) {
            maxTickSize_Hori = index == 0 || index == count - 1 ? outerPadding * 2 : mathAbs(overflowH.tickSpacing.rangeBand) * 2;
        }
    } else if ((helpers.orient == 'left' || helpers.orient == 'right') && (helpers.rotation == 90 || helpers.rotation == 270)) {
        maxTickSize_Vert = index == 0 || index == count - 1 ? outerPadding * 2 : mathAbs(overflowH.tickSpacing.rangeBand) * 2;
    }

    return {
        horizontal: maxTickSize_Hori,
        vertical: maxTickSize_Vert
    };
}

function getBoundaryEdgesForTick(helpers, _bandwidth) {
    let { scale, ticks, rotation, config, chartRect, orient, type } = helpers,
        bandWidth = scale.bandwidth ? scale.bandwidth() / 2 : 0;

    var minTick = ticks[0],
        maxTick = ticks[ticks.length - 1];
    if (config.reversed) {
        (minTick = ticks[ticks.length - 1]), (maxTick = ticks[0]);
    }
    var _bandwidthCopy = pick(_bandwidth, bandWidth);
    var edge = {};
    // var scale1 = axisType == 'x' && hasSplitLeavesColumn && chartObj.axes.x.leaves[0] ? chartObj.axes.x.leaves[0].scale : scaleObj;
    if (fitTextInBetween(helpers)) {
        let distBwTicks = distanceBetweenTicks(helpers),
            outerPadding = mathAbs(distBwTicks.outerPaddingInPixel);

        if (orient == 'bottom' || orient == 'top') {
            (edge.left = scale(minTick) - mathAbs(_bandwidthCopy - outerPadding)),
                (edge.right = scale(maxTick) + (outerPadding + _bandwidthCopy));
        } else if ((orient == 'left' || orient == 'right') && (rotation == 90 || rotation == 270)) {
            (edge.top = scale(maxTick) - outerPadding), (edge.bottom = scale1(minTick) + outerPadding);
        }
    } else {
        edge.top = edge.left = 0;
        edge.right = orient == 'bottom' || orient == 'top' ? helpers.plotSize : 0;
        edge.bottom = orient == 'bottom' || orient == 'top' ? 0 : helpers.plotSize;
    }

    return edge;
}

function getDynamicBoundaryEdgesForTick(node, index, helpers, overflowH, textBounds) {
    let chart = helpers.chart,
        hasDynamicLeafScale =
            chart && chart.dataObject && chart.dataObject.hierarchical && chart.dataObject.hierarchical.dynamicLeafScale,
        edge = {};
    if (overflowH.edgeOptions) {
        edge = overflowH.edgeOptions(node, textBounds);
    } else if (hasDynamicLeafScale && helpers.scale.bandwidth) {
        edge = getBoundaryEdgesForTick(helpers, helpers.scale.bandwidth);
    }
    return edge;
}

function getExtraMarginForBoundary(helpers) {
    var marginPercent = 0.5,
        margin = {},
        labelType = 'ticklabel'; //TODO dynamic label type or add margin
    if (needMinMarginToExcludeFromTrim(helpers)) {
        (margin.left =
            labelType == 'axislabel'
                ? marginPercent * (helpers.chart.canvasarea.width - (helpers.chart.plotarea.left + helpers.chart.plotarea.width))
                : marginPercent * helpers.chart.plotarea.left), //ZC743 If its axislabel margin right and left should be same, because always axislabel placed at center of plotarea with textanchor middle
            (margin.right =
                marginPercent * (helpers.chart.canvasarea.width - (helpers.chart.plotarea.left + helpers.chart.plotarea.width))),
            (margin.top = marginPercent * helpers.chart.plotarea.top),
            (margin.bottom = marginPercent * helpers.chart.plotarea.bottom);
    } else {
        //TODO isPolar
        margin.left = margin.right = 0;
        margin.top = margin.bottom = 0;
    }

    return margin;
}
