// @zohocharts/zc/src/axis/ticks/tick-alignment.js
import { ORDINAL_DATATYPE } from '../../core/datatype-constants';
import { mathAbs } from '../../core/zc-constants';
import { defined } from '../../utils/zcutils';
import { distBetweenRotatedRect } from '../tick-helpers';
import { getTextBound } from '../textBound';

export function tickAlignment(tickSizeH) {
    let _ = this,
        { datatype, rotation, type } = this,
        { ticks, longestTick, tickBound, longestFormattedTick, ticks_formatted } = tickSizeH,
        { tickAlignmentMode, tickRotateMode, tickAutoRotation } = this.rotationConfig,
        tickCount = ticks.length;

    if (tickAlignmentMode === 'auto') {
        //First check for normal ticks with rotation as 0
        var maxCount_Rotate_0 = getMaxTickCountByRotation(0);
        //second check for zigzag only for ordinal datatype, for numeric and time datatype go for rotation with 45
        if (maxCount_Rotate_0.maxTickCount < tickCount) {
            var maxCount_zigzag =
                datatype == ORDINAL_DATATYPE ? getMaxTickCountByZigzag() : getMaxTickCountByRotation(tickAutoRotation, true);
            //third for tick rotation with 45
            if (maxCount_zigzag.maxTickCount < tickCount) {
                var maxCount_Rotate_45 = getMaxTickCountByRotation(tickAutoRotation, true);
                //fourth for ticks with white space
                if (maxCount_Rotate_45.maxTickCount < tickCount && longestTick.hasWhiteSpace) {
                    return getMaxTickCountByWrapOnSpace();
                } else {
                    return maxCount_Rotate_45;
                }
            } else {
                return maxCount_zigzag;
            }
        } else {
            return maxCount_Rotate_0;
        }
    } else if (tickRotateMode) {
        return getMaxTickCountByRotation(null, false);
    } else if (tickAlignmentMode == 'zigzag') {
        return getMaxTickCountByZigzag();
    } else if (tickAlignmentMode == 'wrapOnSpace') {
        return getMaxTickCountByWrapOnSpace();
    } else if (tickAlignmentMode == 'wrapAll') {
        return getMaxTickCountByWrapAll();
    }

    function getMaxTickCountByBounds(rotation, customBounds, includePadding) {
        var tickArea,
            temp,
            padding = 0,
            countForAngle,
            bounds;

        var length = _.plotSize_no_zoom; //getPlotSize(helpers.chart, side, tick.plotRect, gridSize);
        if (includePadding) {
            padding = length * _.chartOuterPadding * 2;
        }

        tickArea = _.plotSize_no_zoom - padding;
        countForAngle = !!rotation;
        bounds = customBounds || tickBound;
        if (countForAngle) {
            if (_.side === 'height' && (rotation === 90 || rotation === 270)) {
                return tickArea / (bounds.width + mathAbs(_.margin.top) + mathAbs(_.margin.bottom));
            }

            var r = rotation % 90;
            var angleCheck = _.side === 'height' ? r > 0 : r > 0 || rotation === 90 || rotation === 270;

            if (angleCheck) {
                var distBwTicks = distBetweenRotatedRect(rotation, bounds.width, bounds.height, _.orient);
                return tickArea / (distBwTicks.tickSpacing + mathAbs(_.margin.left) + mathAbs(_.margin.right));
            }
        }

        if (_.side === 'height') {
            temp = bounds.height + mathAbs(_.margin.top) + mathAbs(_.margin.bottom);
        } else {
            temp = bounds.width + mathAbs(_.margin.left) + mathAbs(_.margin.right);
        }
        return tickArea / temp;
    }

    function getMaxTickCountByRotation(customRotation, recalculate) {
        let rotation_new = defined(customRotation) ? customRotation : rotation;

        var count;
        if (rotation_new === 'auto') {
            rotation = 0;
            count = getMaxTickCountByBounds(0, null, type == 'x');
            var actualRange = _.scale.actualRange;
            var isSingleTick = actualRange && actualRange[0] === actualRange[1];
            if (!isSingleTick && count < tickCount) {
                if (_.side === 'width') {
                    if (_.leafType) {
                        rotation = 90; //For hierarchy auto rotate 90
                    } else {
                        rotation = tickAutoRotation;
                    }

                    count = getMaxTickCountByBounds(tickAutoRotation, null, false); //For auto rotation calculation customType should be null, excludePaddig is true
                    recalculate = true;
                } else {
                    recalculate = true;
                }
            } else {
                recalculate = false;
            }
        } else {
            count = getMaxTickCountByBounds(rotation_new, tickBound, type == 'x');
            rotation = rotation_new;
        }
        return {
            mode: 'rotate',
            rotation: rotation,
            maxTickCount: count,
            recalculate: recalculate
        };
    }

    function getMaxTickCountByInline() {
        var count = getMaxTickCountByBounds(); //For zigzag rotation is 0

        return {
            mode: 'inline',
            rotation: 0,
            maxTickCount: count
        };
    }
    function getMaxTickCountByZigzag() {
        var count = getMaxTickCountByBounds(); //For zigzag rotation is 0

        return {
            mode: 'zigzag',
            rotation: 0,
            maxTickCount: count * 2
        };
    }

    function getMaxTickCountByWrapOnSpace() {
        if (tickAlignmentMode == 'auto') {
            ticks_formatted = ticks_formatted.map(function(d, i) {
                return d.split(' ');
            });
            var longestTick = getLongestTick(
                _.ticks_formatted,
                ticks,
                _.styles.fontSize,
                _.styles.fontFamily,
                _.customTickInfo.styles
            );
            if (!defined(longestFormattedTick)) {
                longestFormattedTick = longestTick.formattedVal;
            }

            // calculate number of tick can be fit.
            var tickBound = getTextBound(
                longestFormattedTick,
                _.styles.fontSize,
                _.styles.fontFamily,
                longestTick.style,
                true
            );
        }

        var count = getMaxTickCountByBounds();

        return {
            mode: 'wrapOnSpace',
            maxTickCount: count,
            rotation: rotation,
            longestFormattedTick
        };
    }

    function getMaxTickCountByWrapAll() {
        var count = getMaxTickCountByBounds();

        return {
            mode: 'wrapAll',
            maxTickCount: count,
            rotation: _.rotation
        };
    }
}

export function getMaxTickCount1(tickCountH, tickH, axisH) {
    let { ticks, tickCount, tickAlignmentMode, tickRotateMode, longestTick, tickBound, tickAutoRotation, rotation } = tickCountH,
        { datatype } = axisH;
    if (tickAlignmentMode === 'auto') {
        //First check for normal ticks with rotation as 0
        var maxCount_Rotate_0 = getMaxTickCountByRotation(tickCountH, tickH, axisH, 0);
        //second check for zigzag only for ordinal datatype, for numeric and time datatype go for rotation with 45
        if (maxCount_Rotate_0.maxTickCount < tickCount) {
            var maxCount_zigzag =
                datatype == ORDINAL_DATATYPE
                    ? getMaxTickCountByZigzag(tickCountH, tickH, axisH)
                    : getMaxTickCountByRotation(tickCountH, tickH, axisH, tickAutoRotation, true);
            //third for tick rotation with 45
            if (maxCount_zigzag.maxTickCount < tickCount) {
                var maxCount_Rotate_45 = getMaxTickCountByRotation(tickCountH, tickH, axisH, tickAutoRotation, true);
                //fourth for ticks with white space
                if (maxCount_Rotate_45.maxTickCount < tickCount && longestTick.hasWhiteSpace) {
                    return getMaxTickCountByWrapOnSpace(tickCountH, tickH, axisH);
                } else {
                    return maxCount_Rotate_45;
                }
            } else {
                return maxCount_zigzag;
            }
        } else {
            return maxCount_Rotate_0;
        }
    } else if (tickRotateMode) {
        return getMaxTickCountByRotation(tickCountH, tickH, axisH, null, false);
    } else if (tickAlignmentMode == 'zigzag') {
        return getMaxTickCountByZigzag(tickCountH, tickH, axisH);
    } else if (tickAlignmentMode == 'wrapOnSpace') {
        return getMaxTickCountByWrapOnSpace(tickCountH, tickH, axisH);
    } else if (tickAlignmentMode == 'wrapAll') {
        return getMaxTickCountByWrapAll(tickCountH, tickH, axisH);
    }
}

function getMaxTickCountByBounds(tickCountH, tickH, axisH, rotation, customBounds, includePadding) {
    let { ticks, tickCount, tickAlignmentMode, tickRotateMode, longestTick, tickBound, tickAutoRotation } = tickCountH,
        { dataType, type, plotRect, orient } = axisH,
        { side, margin, chartOuterPadding, plotSize_no_zoom, size } = tickH;
    var tickArea,
        temp,
        padding = 0,
        countForAngle,
        bounds;

    var length = plotSize_no_zoom; //getPlotSize(helpers.chart, side, tick.plotRect, gridSize);
    if (includePadding) {
        padding = length * chartOuterPadding * 2;
    }

    tickArea = plotSize_no_zoom - padding;
    countForAngle = !!rotation;
    bounds = customBounds || tickBound;
    if (countForAngle) {
        if (side === 'height' && (rotation === 90 || rotation === 270)) {
            return tickArea / (bounds.width + mathAbs(margin.top) + mathAbs(margin.bottom));
        }

        var r = rotation % 90;
        var angleCheck = side === 'height' ? r > 0 : r > 0 || rotation === 90 || rotation === 270;

        if (angleCheck) {
            var distBwTicks = distBetweenRotatedRect(rotation, bounds.width, bounds.height, orient);
            return tickArea / (distBwTicks.tickSpacing + mathAbs(margin.left) + mathAbs(margin.right));
        }
    }

    if (side === 'height') {
        temp = bounds.height + mathAbs(margin.top) + mathAbs(margin.bottom);
    } else {
        temp = bounds.width + mathAbs(margin.left) + mathAbs(margin.right);
    }
    return tickArea / temp;
}

function getMaxTickCountByRotation(tickCountH, tickH, axisH, customRotation, recalculate) {
    var { rotation, side, leafType } = tickH,
        { dataType, type } = axisH,
        { ticks, tickCount, tickAlignmentMode, tickRotateMode, longestTick, tickBound, tickAutoRotation, rotation } = tickCountH,
        rotation_new = defined(customRotation) ? customRotation : rotation;

    var count;
    if (rotation_new === 'auto') {
        rotation = 0;
        count = getMaxTickCountByBounds(tickCountH, tickH, axisH, 0, null, type == 'x');
        var actualRange = tickH.scale.actualRange;
        var isSingleTick = actualRange && actualRange[0] === actualRange[1];
        if (!isSingleTick && count < tickCount) {
            if (side === 'width') {
                if (leafType) {
                    rotation = 90; //For hierarchy auto rotate 90
                } else {
                    rotation = tickAutoRotation;
                }
                // isBranchAxisType = chart.dataObject.pivot && type == 'x' && isBranchAxis(type, axisOrder, chart),
                // if (
                //     !tickCountH.hasDefinedAutoRotation &&
                //     ((categoryHierarchy && type == 'x' && !isBranchAxisType && hasBranchAxis(type, helpers.chart)) ||
                //         (!categoryHierarchy && type == 'x' && helpers.chart.axes.x.count > 1))
                // ) {
                //     rotation = 90; //For hierarchy auto rotate 90
                // } else {
                //
                // }

                count = getMaxTickCountByBounds(tickCountH, tickH, axisH, tickAutoRotation, null, false); //For auto rotation calculation customType should be null, excludePaddig is true
                recalculate = true;
            } else {
                recalculate = true;
            }
        } else {
            recalculate = false;
        }
    } else {
        count = getMaxTickCountByBounds(tickCountH, tickH, axisH, rotation_new, tickBound, type == 'x');
        rotation = rotation_new;
    }
    return {
        mode: 'rotate',
        rotation: rotation,
        maxTickCount: count,
        recalculate: recalculate
    };
}

function getMaxTickCountByInline(tickCountH, tickH, axisH) {
    var count = getMaxTickCountByBounds(tickCountH, tickH, axisH); //For zigzag rotation is 0

    return {
        mode: 'inline',
        rotation: 0,
        maxTickCount: count
    };
}
function getMaxTickCountByZigzag(tickCountH, tickH, axisH) {
    var count = getMaxTickCountByBounds(tickCountH, tickH, axisH); //For zigzag rotation is 0

    return {
        mode: 'zigzag',
        rotation: 0,
        maxTickCount: count * 2
    };
}

function getMaxTickCountByWrapOnSpace(tickCountH, tickH, axisH) {
    let {
            ticks,
            ticks_formatted,
            tickCount,
            tickAlignmentMode,
            tickRotateMode,
            tickAutoRotation,
            rotation,
            customTickInfo,
            longestFormattedString
        } = tickCountH,
        { styles } = ticks;
    if (tickAlignmentMode == 'auto') {
        ticks_formatted = ticks_formatted.map(function(d, i) {
            return d.split(' ');
        });
        var longestTick = getLongestTick(ticks_formatted, ticks, styles.fontSize, styles.fontFamily, customTickInfo.styles);
        if (!defined(longestFormattedString)) {
            longestFormattedString = longestTick.formattedVal;
        }

        // calculate number of tick can be fit.
        var tickBound = getTextBound(longestFormattedString, styles.fontSize, styles.fontFamily, longestTick.style, true);
    }

    var count = getMaxTickCountByBounds(tickCountH, tickH, axisH);

    return {
        mode: 'wrapOnSpace',
        maxTickCount: count,
        rotation: rotation
    };
}

function getMaxTickCountByWrapAll(tickCountH, tickH, axisH) {
    var count = getMaxTickCountByBounds(tickCountH, tickH, axisH);

    return {
        mode: 'wrapAll',
        maxTickCount: count,
        rotation: tickCountH.rotation
    };
}
