// @zohocharts/zc/src/axis/radial/parser/config/radial-renderer-config.js
import { LEFT_ALIGN, RIGHT_ALIGN } from "../../../../core/align-constants";
import { d3_arc } from "../../../../core/d3-constants";
import { mathAbs } from "../../../../core/zc-constants";
import { polarToXY } from "../../../../polaraxis/polarutils";
import { defined, isArray, pick, validateNullVal } from "../../../../utils/zcutils";
import { getAxisObject } from "../../../../polaraxis/polaraxis-helpers";
import { getTickRotateInfo } from "../../../tickRotate-helpers";

export function getRadialCommonOption() {
    const self = this;

    const { renderer, axis, chart, data } = this,
        minorGrid = null,   // no minor grid for radial axis
        isLevelType = axis.ticks.isHierarchialLevelType(),
        { tickmark: tickmarkdata, subtickmark: subtickmarkdata = {} } = data;

    // calculate values from renderer here to reduce arguments in parser
    const yaxis_data = chart.systemConf.chart.axes.yaxis[0],
        orient = yaxis_data.orient || RIGHT_ALIGN,
        tickSizeOuter = validateNullVal(yaxis_data.tickmark.size, 6),
        tickSizeInner = tickSizeOuter,
        tickPadding = validateNullVal(yaxis_data.ticklabel[orient == LEFT_ALIGN ? 'marginRight' : 'marginLeft'], 3);

    const tickObj = axis.ticks.tickValueObj,
        isInline = tickObj?.isInline,
        isInwardPlot = tickObj?.inlineToPlot,
        k = orient === 'top' || orient === 'left' ? -1 : 1,
        l = isInwardPlot ? -k : k,
        m = isInline ? -k : k,
        x = orient === 'left' || orient === 'right' ? 'x' : 'y',
        w = orient === 'left' || orient === 'right' ? 'width' : 'height',
        y = x == 'x' ? 'y' : 'x',
        h = chart.dataObject.isPolarAxisCategory ? 'H' : 'h';

    const tickConstants = { k, l, m, x, w, y, h };

    const tickRotation = pick(renderer.tickRotation(), 0),
        rotation = tickRotation < 0 ? 360 - mathAbs(tickRotation) : tickRotation,
        hasMinorGridPositions = isArray(minorGrid) && minorGrid.length > 0,
        isSubTickMarkShow = (subtickmarkdata.count > 0 || defined(subtickmarkdata.interval)) && hasMinorGridPositions,
        subTickSize = isSubTickMarkShow ? renderer.subTickSize?.() : 0,
        rotationInfo = getTickRotateInfo(orient, rotation, null, null, Math.max(tickmarkdata.size, subTickSize), tickPadding, tickObj);


    const tickmarkStart = isInline && !isInwardPlot ? tickObj.axisgroup[w] : 0,
        transform = orient === 'top' || orient === 'bottom' ? translateX : translateY,
        scale1 = renderer.scale(),
        range = scale1.origRange ? scale1.origRange : scale1.range(),
        range0 = +range[0] + 0.5,
        range1 = +range[range.length - 1] + 0.5,
        position = (scale1.bandwidth ? (scale1.discreteWidth ? (isLevelType ? number$1 : center1) : center) : number$1)(scale1.copy());

    const spacing = isInline && !isInwardPlot ? tickObj.longTickSize[w] : Math.max(tickSizeInner, subTickSize, 0) + tickPadding;

    const commonOption = {
        isInline, isInwardPlot, tickConstants,
        tickmarkStart, tickSizeOuter, tickSizeInner, tickPadding,
        range, range0, range1, orient, transform, position,
        spacing, rotationInfo
    };
    return commonOption;

}

export function getRadialComponent() {
    const self = this;

    const { renderer, axis, config, chart } = this,
        { grid } = config,
        tickObj = axis.ticks.tickValueObj,
        renderTicklabel = tickObj.inlineToPlot ? false : true,
        isGridShow = grid.show,
        renderGrid = !tickObj.inlineToPlot,
        xscale = getAxisObject(chart, 'x', 0).scale;

    const component = {
        scale: renderer.scale(),
        xscale,
        gridPath,
        renderTicklabel,
        isGridShow, renderGrid,
        tickFormat: renderer.tickFormat(),
        tickValues: renderer.tickValues(),
        radius: renderer.radius()
    };

    return component;
}

// add dependent functions here
function translateX(x) {
    return `translate(${x + 0.5},0)`;
}

function translateY(y) {
    return `translate(0,${y + 0.5})`;
}

function number$1(scale1) {
    return function (d) {
        return +scale1(d);
    };
}

function center(scale1) {
    var offset = Math.max(0, scale1.bandwidth() - 1) / 2; // Adjust for 0.5px offset.
    if (scale1.round()) {
        offset = Math.round(offset);
    }
    return function (d) {
        return +scale1(d) + offset;
    };
}

function center1(scale1) {
    return function (d, i) {
        var offset = Math.max(0, scale1.bandwidth(d) - 1) / 2; // Adjust for 0.5px offset.
        // if (scale1.round()) {
        offset = Math.round(offset);
        // }
        return +scale1(d) + offset;
    };
}

function gridPath(r, scaleObj, axisConfig) {
    //TODO: grid customization here @SH
    let range = scaleObj.range();
    if (axisConfig.grid.type === 'arc') {
        var path = d3_arc()
            .innerRadius(r - 1)
            .outerRadius(r)
            .startAngle(range[0])
            .endAngle(range[1])(1);
        return `M${path.split(/L|M/gi)[1]}`;
    } else {
        var tickValues = scaleObj.ticks ? scaleObj.ticks() : scaleObj.domain();
        var path = tickValues.reduce(function (prev, cur, i) {
            var a = scaleObj(cur);
            var xy = polarToXY(r, a);
            return `${prev}L${xy[0]} ${xy[1]} `;
        }, '');
        return `M${path.substring(1)}Z`;
    }
}
