// @zohocharts/zc/src/label/bardatalabels.js
//$Id$
//bardatalabels.js
import { LabelRegistry } from './label-registry';
import { getPlotBound } from '../plot/plotSize';
import { DataProcessor } from '../core/data-processor';
import { DATALABELS_LABELTYPE, STACKLABELS_LABELTYPE } from './labeltype-constants';
import {
    getMergedDatalabelOptions,
    applyDatalabelFonts,
    callTrimText,
    isLabelWithinBound,
    isDatalabelEnabled,
    isStacklabelEnabled,
    getLabelArray
} from './datalabelsUtils';
import { NULL } from '../core/zc-constants';
import { isMultiColoring, isWithinColorRange, isSeriesFiltered } from '../utils/colorutils';
import {
    FIXEDTOP_ALIGN,
    HORIZONTAL_ALIGN,
    VERTICAL_ALIGN,
    TOP_ALIGN,
    LEFT_ALIGN,
    RIGHT_ALIGN,
    MIDDLE_ALIGN,
    BOTTOM_ALIGN
} from '../core/align-constants';
import {
    getMergedBoundBoxOptions,
    isBoundBoxEnabled,
    getBoundBoxStrokeWidth,
    getBoundBoxPadding,
    getBoundBoxHookConf
} from './boundBoxUtils';
import { validateNullVal, defined, bbox, toArray, perToPx, pick } from '../utils/zcutils';
import { TEXT_ELEMENT } from '../core/element-constants';
import { translateTspan, getTspanBound, trimAllTspan } from './tspan-utils';
import { getSizeToBeAdded } from './boundBoxUtils';
import { ORDINAL_DATATYPE } from '../core/datatype-constants';
import { d3_max, d3_set } from '../core/d3-constants';
import { mathMax, mathMin, NONE } from '../core/zc-constants';
import { BOXPLOT_CHARTNAME, BAR_CHARTNAME, BARRANGE_CHARTNAME, BULLET_CHARTNAME } from '../core/chartname-constants';
import { getFontSizeByHierarchy } from '../utils/zcdatautils';
import { appendEle } from '../utils/zcsvgutils';
import { dataLabelTextContent } from './datalabel-content';
import { getValignforBarLabels } from './checkFixedTopLabels';
import { getDy4MultiLine } from './position-helpers';
import { getChartTypeByIndex } from '../core/charttype-helpers';

/*
 * PENDING:
 * *Bound box for stack labels
 */
function barLabel() {
    var renderer, args;
    var axisrotated,
        textDirection,
        rotation,
        dataProcessorHelpers = DataProcessor.helpers;

    function labels(selection, labelHandler) {
        selection.each(function () {
            var currentdata = args[0],
                index = args[1],
                labelType = args[2] || DATALABELS_LABELTYPE,
                yrangePos = args[3],
                columnindex = args[4],
                rendererIndex = pick(labelHandler.rendererIndex, index),
                paneid = labelHandler.paneid,
                rp = renderer.commonRendererProp().params,
                dataindex = defined(columnindex) ? rp.chartObj.systemConf.metadata.columns[columnindex].dataindex : null,
                plotoptions = currentdata.plotoptions,
                dataLabels,
                chartObj = rp.chartObj,
                systemconf = chartObj.systemConf,
                commonPlotOptions = rp.commonPlotOptions,
                dataset = rp.dataset,
                stacked = rp.stacked,
                yscale = rp.yscale,
                yaxiscolumnorder = rp.yaxiscolumnorder,
                labelArray = getLabelArray(rp, paneid),
                isStackedLabels = labelType === STACKLABELS_LABELTYPE,
                cType_Obj = getChartTypeByIndex(chartObj, paneid, index),
                isRangeChartType = cType_Obj.name == BOXPLOT_CHARTNAME || cType_Obj.name == BARRANGE_CHARTNAME, //consider median for boxplot
                hasRangePos = isRangeChartType && defined(yrangePos), //consider median for boxplot
                isCanvasMode = chartObj.plot.renderer.mode === 'canvas',
                axisrotated = rp.axisrotated;

            let renderWithinPlot = true,
                renderOnInit = chartObj.systemConf.chart.plot.datalabels.renderOnInit;
            
            const isComboChart = chartObj.dataObject.isCombinationalChart;

            // #ZC2228
            if (!renderOnInit) {
                var behaviours = chartObj.behaviouralEvents.eventManager.behaviours
                if (behaviours.PLOT_ZOOM || (behaviours.CHART_SCROLL && !axisrotated)) {
                    renderWithinPlot = false;
                }
            }

            dataLabels = getMergedDatalabelOptions(commonPlotOptions, isStackedLabels ? NULL : plotoptions, chartObj, index, paneid);

            if (isStackedLabels) {
                dataLabels = dataLabels.stackLabels;
            }
            var showlabel = isStackedLabels ? isStacklabelEnabled(dataLabels, chartObj) : isDatalabelEnabled(dataLabels, chartObj);

            if (currentdata.disabled || !showlabel || (!isStackedLabels && isSeriesFiltered(chartObj, paneid, index))) {
                return;
            }

            var currentPlotBound = getPlotBound(chartObj);
            var textOverlap = dataLabels.textOverlap, //hidden | adjust | clip
                labelPos = getValignforBarLabels(chartObj, index, labelType), //bottom | top | middle
                showZeros = dataLabels.showZeros,
                multiColoring = isMultiColoring(commonPlotOptions, chartObj),
                left = currentPlotBound.left,
                top = currentPlotBound.top,
                right = currentPlotBound.right,
                bottom = currentPlotBound.bottom,
                rectBound = chartObj.svg.node().getBoundingClientRect(),
                rectBoundLeft = rectBound.left + chartObj.plotarea.left,
                rectBoundTop = rectBound.top + chartObj.plotarea.top,
                dim = (axisrotated ? chartObj.plotarea.width : chartObj.plotarea.height) * 0.75,
                GLT = dataLabelTextContent(chartObj, rp, dataLabels, labelType, NULL, index, yrangePos, paneid),
                contentType = GLT.contentType || 'text',
                isMultiLine = contentType === 'html' && GLT.multiLine,
                isFixedTopPosition = labelPos === FIXEDTOP_ALIGN,
                isTopPosition = labelPos === TOP_ALIGN,
                isCatNullX = dataProcessorHelpers.isCatNull(dataset, 'x'),
                isCatNullY = dataProcessorHelpers.isCatNull(dataset, 'y', yaxiscolumnorder),
                yaxisReversed = systemconf.chart.axes.yaxis[yaxiscolumnorder || 0].reversed,
                barW = chartObj.barscale2.bandwidth() || 0,
                fontSize = getFontSizeByHierarchy(dataLabels, chartObj),
                datatypeY = chartObj.dataset.getDataType('y', yaxiscolumnorder);

            // if (isFixedTopPosition && !chartObj.cache.correction['label-y-' + yaxiscolumnorder]) {
            //     return;
            // }

            var boundBoxPadding = 0,
                boundBoxStrokeWidth = 0,
                hookHeight = 0,
                hookWidth = 0,
                boundBoxRadius = 0;
            if (isBoundBoxEnabled(dataLabels, chartObj)) {
                var boundBoxOptions = getMergedBoundBoxOptions(dataLabels, chartObj);
                var hookOptions = getBoundBoxHookConf(boundBoxOptions.hook);
                boundBoxPadding = getBoundBoxPadding(dataLabels, chartObj);
                boundBoxStrokeWidth = getBoundBoxStrokeWidth(dataLabels, chartObj);
                hookHeight = stacked || labelPos === MIDDLE_ALIGN ? 0 : hookOptions.height;
                hookWidth = hookOptions.width;
                boundBoxRadius = boundBoxOptions.radius;
            }

            var clip = textOverlap === 'clip',
                handleOverlapping = systemconf.chart.plot.datalabels.handleOverlapping && textOverlap !== NONE,
                padding =
                    mathMax(0, validateNullVal(dataLabels.padding, 2)) +
                    boundBoxPadding +
                    hookHeight +
                    (isFixedTopPosition && boundBoxStrokeWidth ? boundBoxStrokeWidth / 2 - 1 : 0),
                fixedTopPosition = isFixedTopPosition && getFixedTopPosition();
            padding = padding > dim ? dim : padding;
            textDirection = dataLabels.textDirection;
            textDirection =
                !textDirection || (stacked && !isStackedLabels) ? (axisrotated ? VERTICAL_ALIGN : HORIZONTAL_ALIGN) : textDirection;
            rotation = axisrotated ? (textDirection == HORIZONTAL_ALIGN ? 90 : 0) : textDirection == HORIZONTAL_ALIGN ? 0 : 270;
            var commonOption = renderer.commonOption;
            var getParamValue = function () {
                return chartObj.dataObject.getTransRectValue.apply(chartObj.dataObject, toArray(arguments));
            };

            var temp_label = appendEle(chartObj.staticPaneSVG.select('.labelholder'), TEXT_ELEMENT, [1], '.', 'class', 'temp_label');
            var updateHandler = labelHandler.update,
                queue = labelHandler.add;
            if (chartObj.viewFlow && chartObj.viewFlow.marks._default.label.length) {
                let marks_plot = chartObj.viewFlow.marks._default.label[index];
                currentdata = marks_plot.data;
            }
            currentdata.data.forEach(function (d, i) {
                var dataUpdate =
                    renderer.seriesdataAfterUpdate && !isStackedLabels ? renderer.seriesdataAfterUpdate[i] : currentdata.data[i];

                if (isStackedLabels) {
                    var scrollOptions = systemconf.chart.scroll;
                    if (scrollOptions && scrollOptions.enabled) {
                        // #ZC1374
                        dataUpdate = getVisibleData(dataUpdate, rp.xscale.domain());
                    }

                    if (isFixedTopPosition && yscale.domain()[0] < 0) {
                        dataUpdate = getUniqueDataByX(dataUpdate);
                    }
                }

                var commonProperties = [];
                applyDatalabelFonts(temp_label, dataLabels, chartObj, rp.color, false, index);
                temp_label.style('opacity', 0);

                dataUpdate.forEach(function (d, j) {
                    var text,
                        boundBox = NULL,
                        display = true,
                        flip = yaxisReversed ? -1 : 1,
                        isNeg = isStackedLabels ? d[1] < 0 : dataset.getY(d, yaxiscolumnorder) < 0;

                    temp_label[contentType](function () {
                        if (isStackedLabels) {
                            var yVal = d[1],
                                xVal = d[0];
                            text = !showZeros && yVal == 0 ? NULL : GLT(d, j);
                        } else if (
                            isCatNullX(dataset.getX(d)) ||
                            isCatNullY(dataset.getY(d, yaxiscolumnorder)) ||
                            (!showZeros && dataset.getY(d, yaxiscolumnorder) == 0)
                        ) {
                            text = NULL;
                        } else {
                            text = GLT(d, j);
                        }
                        return text;
                    });

                    if (isMultiLine) {
                        translateTspan(temp_label, 0);
                    }

                    var X, Y, DY, ANCHOR, HOOKPOSITION, actualDY;
                    boundBox = bbox(temp_label.node());
                    var widthToBeAdded = 0,
                        heightToBeAdded = 0;

                    if (isStackedLabels) {
                        // boundBox = bbox(temp_label.node());
                        var x = commonOption.bandwidth1(d[0]) + commonOption.bandwidth2(commonOption.stackGroupIndex), //ZC1419
                            y = yscale(d[2] || d[1]); //For waterfall charts, there will be cascaded value
                        X = axisrotated ? y : x;
                        Y = axisrotated ? x : y;
                        if (axisrotated) {
                            if (textDirection == HORIZONTAL_ALIGN) {
                                X += (isNeg ? -padding : padding) * flip;
                                DY = yaxisReversed ? (isNeg ? '-0.3em' : '1em') : isNeg ? '1em' : '-0.3em';
                            } else {
                                Y += barW / 2;
                                DY = '0.3em';
                                if (isFixedTopPosition) {
                                    X = fixedTopPosition + padding * flip + (isMultiLine ? 0 : 0) * flip;
                                    ANCHOR = yaxisReversed ? 'end' : NULL;
                                } else {
                                    X += (isNeg ? -padding : padding) * flip;
                                    ANCHOR = yaxisReversed ? (isNeg ? NULL : 'end') : isNeg ? 'end' : NULL;

                                    if (isMultiLine) {
                                        X += (boundBox.width / 2) * (isNeg ? -1 : 1) * flip;
                                    }
                                }
                            }
                        } else {
                            X += barW / 2;
                            Y += (isNeg ? padding : -padding) * flip;
                            if (textDirection == VERTICAL_ALIGN) {
                                DY = '0.3em';
                                ANCHOR = yaxisReversed ? (isNeg ? NULL : 'end') : isNeg ? 'end' : NULL;
                            } else {
                                ANCHOR = 'middle';
                                if (isFixedTopPosition) {
                                    Y = fixedTopPosition - padding * flip;
                                    DY = yaxisReversed ? '1em' : '-0.3em';
                                } else {
                                    DY = yaxisReversed ? (isNeg ? '-0.3em' : '1em') : isNeg ? '1em' : '-0.3em';
                                }
                            }
                        }
                    } else {
                        // var b = boundBox = bbox(temp_label.node()),
                        var b = boundBox,
                            x =
                                defined(dataindex) && axisrotated
                                    ? rp.yscale(d[dataindex][yrangePos] || d[dataindex])
                                    : getParamValue('x', d, i, j, commonOption, null, yrangePos) || 0,
                            y =
                                defined(dataindex) && !axisrotated
                                    ? rp.yscale(d[dataindex][yrangePos] || d[dataindex])
                                    : getParamValue('y', d, i, j, commonOption, null, yrangePos) || 0, //consider median for boxplot {d[dataindex] is not an array}
                            width = getParamValue('width', d, i, j, commonOption) || 0,
                            height = getParamValue('height', d, i, j, commonOption) || 0,
                            CLIP = clip;
                        X = x;
                        Y = y;
                        var hasNegRange = hasRangePos && defined(dataindex) ? d[dataindex][0] > d[dataindex][1] : null,
                            isInverted = yaxisReversed ? (hasNegRange ? true : false) : hasNegRange ? true : false,
                            isInverted_yAlign = yaxisReversed && axisrotated ? !isInverted : isInverted,
                            curYalignPos = isInverted_yAlign ? 1 - yrangePos : yrangePos, //For negRange invert yrangePos
                            curYHookPos = yaxisReversed
                                ? hasNegRange
                                    ? yrangePos
                                    : 1 - yrangePos
                                : hasNegRange
                                    ? 1 - yrangePos
                                    : yrangePos,
                            flip = isRangeChartType && hasRangePos ? (yaxisReversed && !axisrotated ? -1 : 1) : flip; //NEED TO CHECK ZC2039

                        if (axisrotated) {
                            if (stacked) {
                                Y += height / 2;
                                DY = '0.3em';
                                if (labelPos == MIDDLE_ALIGN) {
                                    X += (width / 2) * flip;
                                    ANCHOR = 'middle';
                                } else if (labelPos == TOP_ALIGN) {
                                    X += (isNeg ? padding : width - padding) * flip;
                                    ANCHOR = yaxisReversed ? (isNeg ? 'end' : NULL) : isNeg ? NULL : 'end';

                                    if (isMultiLine && !isFixedTopPosition) {
                                        X += (b.width / 2) * (isNeg ? 1 : -1) * flip;
                                    }
                                } else {
                                    X += (isNeg ? width - padding : padding) * flip;
                                    ANCHOR = yaxisReversed ? (isNeg ? NULL : 'end') : isNeg ? 'end' : NULL;

                                    if (isMultiLine && !isFixedTopPosition) {
                                        X += (b.width / 2) * (isNeg ? -1 : 1) * flip;
                                    }
                                }
                            } else {
                                if (textDirection == VERTICAL_ALIGN) {
                                    heightToBeAdded = getSizeToBeAdded(boundBox.height, boundBoxPadding, boundBoxRadius, hookWidth);
                                    Y += height / 2 + heightToBeAdded / 2;
                                    DY = '0.3em';
                                    CLIP = CLIP && labelPos != TOP_ALIGN && b.width > width;
                                    HOOKPOSITION = yaxisReversed ? (isNeg ? LEFT_ALIGN : RIGHT_ALIGN) : isNeg ? RIGHT_ALIGN : LEFT_ALIGN;

                                    if (isFixedTopPosition) {
                                        ANCHOR = yaxisReversed ? 'end' : NULL;
                                        X = fixedTopPosition + padding * flip + (isMultiLine ? 0 : 0) * flip;
                                        HOOKPOSITION = yaxisReversed ? RIGHT_ALIGN : LEFT_ALIGN;
                                    } else if (!CLIP) {
                                        ANCHOR = yaxisReversed ? (isNeg ? NULL : 'end') : isNeg ? 'end' : NULL;
                                        if (labelPos == MIDDLE_ALIGN) {
                                            X += (b.width > width ? (isNeg ? width : 0) : width / 2) * flip;
                                            ANCHOR = b.width > width ? ANCHOR : 'middle';
                                        } else if (labelPos == TOP_ALIGN) {
                                            if (isRangeChartType) {
                                                if (hasRangePos) {
                                                    X += (curYalignPos == 0 ? -padding : padding) * flip;
                                                    ANCHOR = curYalignPos == 0 ? 'end' : NULL;
                                                    HOOKPOSITION = curYHookPos == 0 ? RIGHT_ALIGN : LEFT_ALIGN;
                                                } else {
                                                    // X += (b.width > width ? (isNeg ? width : 0) : width / 2) * flip;
                                                    ANCHOR = 'middle';
                                                    HOOKPOSITION = NULL;
                                                }
                                            } else {
                                                X += (isNeg ? -padding : width + padding) * flip;
                                            }
                                            // X += (isNeg ? -padding : width + padding) * flip;
                                        } else {
                                            X += (isNeg ? width - padding : padding) * flip;
                                        }

                                        if (isMultiLine && !(labelPos === MIDDLE_ALIGN && b.width <= width)) {
                                            X += (isNeg ? -boundBox.width / 2 : boundBox.width / 2) * flip;
                                        }
                                    } else {
                                        if (labelPos == BOTTOM_ALIGN) {
                                            X += yaxisReversed ? (isNeg ? -width + padding : -width) : isNeg ? 0 : padding;
                                        } else if (labelPos == MIDDLE_ALIGN) {
                                            X += yaxisReversed ? -width + boundBoxPadding : boundBoxPadding;
                                        }
                                    }
                                } else {
                                    widthToBeAdded = getSizeToBeAdded(boundBox.width, boundBoxPadding, boundBoxRadius, hookWidth);
                                    DY = yaxisReversed ? (isNeg ? '-0.3em' : '1em') : isNeg ? '1em' : '-0.3em';
                                    CLIP = CLIP && b.width > height;
                                    HOOKPOSITION = yaxisReversed ? (isNeg ? BOTTOM_ALIGN : TOP_ALIGN) : isNeg ? TOP_ALIGN : BOTTOM_ALIGN;

                                    if (labelPos == MIDDLE_ALIGN) {
                                        if (b.height > width) {
                                            if (isMultiLine) {
                                                X += (yaxisReversed ? (isNeg ? width : 0) : isNeg ? 0 : width) * flip;
                                                X += yaxisReversed ? (b.height / 2) * (isNeg ? 1 : -1) : 0;
                                            } else {
                                                X += (yaxisReversed ? (isNeg ? 0 : width) : isNeg ? width : 0) * flip;
                                            }
                                        } else {
                                            X += (width / 2) * flip;
                                        }
                                        DY = b.height > width ? DY : '0.3em';
                                    } else if (labelPos == TOP_ALIGN) {
                                        if (isRangeChartType) {
                                            if (hasRangePos) {
                                                X += (curYalignPos == 0 ? -padding : padding) * flip;
                                                ANCHOR = curYalignPos == 0 ? 'end' : NULL;
                                                HOOKPOSITION = curYHookPos == 0 ? RIGHT_ALIGN : LEFT_ALIGN;
                                                DY = curYHookPos == 0 ? '1em' : '-0.3em';
                                                // DY = yaxisReversed ? (isNeg ? '-0.3em' : '1em') : isNeg ? '1em' : '-0.3em';
                                            } else {
                                                // X += (b.width > width ? (isNeg ? width : 0) : width / 2) * flip;
                                                ANCHOR = 'middle';
                                                HOOKPOSITION = NULL;
                                            }
                                        } else {
                                            X += (isNeg ? -padding : width + padding) * flip;
                                        }
                                    } else {
                                        X += (isNeg ? width - padding : padding) * flip;
                                    }
                                    if (!CLIP) {
                                        Y += height / 2;
                                        ANCHOR = 'middle';
                                        if (labelPos == MIDDLE_ALIGN) {
                                            DY = '0.3em';
                                        }
                                    } else {
                                        X += widthToBeAdded / 2;
                                        Y += boundBoxPadding;
                                    }
                                }
                            }
                        } else {
                            if (stacked) {
                                X += width / 2;
                                ANCHOR = 'middle';
                                if (labelPos == MIDDLE_ALIGN) {
                                    Y += (height / 2) * flip;
                                    DY = '0.3em';
                                } else if (labelPos == TOP_ALIGN) {
                                    Y += (isNeg ? height - padding : padding) * flip;
                                    DY = yaxisReversed ? (isNeg ? '1em' : '-0.3em') : isNeg ? '-0.3em' : '1em';
                                } else {
                                    Y += (isNeg ? padding : height - padding) * flip;
                                    DY = yaxisReversed ? (isNeg ? '-0.3em' : '1em') : isNeg ? '1em' : '-0.3em';
                                }
                            } else {
                                if (textDirection == VERTICAL_ALIGN) {
                                    heightToBeAdded = getSizeToBeAdded(boundBox.height, boundBoxPadding, boundBoxRadius, hookWidth);
                                    var BW = (!yaxisReversed && isNeg) || (yaxisReversed && !isNeg) ? b.width : 0;
                                    CLIP = CLIP && labelPos != TOP_ALIGN && b.width + padding > height;
                                    DY = '0.3em';
                                    HOOKPOSITION = yaxisReversed ? (isNeg ? LEFT_ALIGN : RIGHT_ALIGN) : isNeg ? RIGHT_ALIGN : LEFT_ALIGN;

                                    if (!CLIP) {
                                        X += width / 2;
                                        if (labelPos == MIDDLE_ALIGN) {
                                            if (b.width > height) {
                                                if (isMultiLine) {
                                                    Y += (isNeg ? boundBox.width / 2 : height - boundBox.width / 2) * flip;
                                                } else {
                                                    Y += (isNeg ? 0 : height) * flip + BW;
                                                }
                                            } else {
                                                Y += (height / 2) * flip;
                                            }
                                            ANCHOR = b.width > height ? NULL : 'middle';
                                        } else if (labelPos == TOP_ALIGN) {
                                            Y += (isNeg ? height + padding : -padding) * flip + BW;
                                            if (isMultiLine) {
                                                Y -= boundBox.width / 2;
                                            }
                                        } else {
                                            Y += (isNeg ? padding : height - padding) * flip + BW;
                                            if (isMultiLine) {
                                                Y -= boundBox.width / 2;
                                            }
                                        }
                                    } else {
                                        X += width / 2;
                                        if (labelPos == MIDDLE_ALIGN) {
                                            Y += yaxisReversed ? boundBoxPadding : height - boundBoxPadding;
                                        } else {
                                            Y += yaxisReversed ? (isNeg ? -padding : 0) : height + (isNeg ? 0 : -padding);
                                        }
                                    }
                                } else {
                                    widthToBeAdded = getSizeToBeAdded(boundBox.width, boundBoxPadding, boundBoxRadius, hookWidth);
                                    CLIP = CLIP && b.width + 2 * boundBoxPadding > width;
                                    DY = yaxisReversed
                                        ? isNeg || curYalignPos == 0
                                            ? '-0.3em'
                                            : '1em'
                                        : isNeg || curYalignPos == 0
                                            ? '1em'
                                            : '-0.3em';
                                    HOOKPOSITION = yaxisReversed ? (isNeg ? BOTTOM_ALIGN : TOP_ALIGN) : isNeg ? TOP_ALIGN : BOTTOM_ALIGN;

                                    if (isFixedTopPosition) {
                                        Y = fixedTopPosition - padding * flip;
                                        HOOKPOSITION = yaxisReversed ? TOP_ALIGN : BOTTOM_ALIGN;
                                        DY = yaxisReversed ? '1em' : '-0.3em';
                                    } else if (labelPos == MIDDLE_ALIGN) {
                                        Y += (b.height < height ? height / 2 : isNeg ? 0 : height) * flip;
                                        DY = b.height < height ? '0.3em' : DY;
                                    } else if (labelPos == TOP_ALIGN) {
                                        Y += (isNeg && !hasRangePos ? height + padding : curYalignPos == 0 ? padding : -padding) * flip;
                                        HOOKPOSITION = isRangeChartType
                                            ? hasRangePos
                                                ? curYHookPos == 0
                                                    ? TOP_ALIGN
                                                    : BOTTOM_ALIGN
                                                : NULL
                                            : HOOKPOSITION;
                                    } else {
                                        Y += (isNeg ? padding : height - padding) * flip;
                                    }
                                    if (!CLIP) {
                                        X += width / 2;
                                        ANCHOR = 'middle';
                                    } else {
                                        X += boundBoxPadding;
                                    }
                                }
                            }
                        }
                    }

                    if (isMultiLine) {
                        let t = getDy4MultiLine(temp_label, dataLabels, parseFloat(DY), text, display);
                        DY = `${t[0]}em`;
                        actualDY = t[1]; //check
                        text = t[2];
                        display = t[3];

                        if (!CLIP && !isFixedTopPosition) {
                            ANCHOR = 'middle';
                        }
                    }

                    temp_label.attrs({
                        x: X,
                        y: Y,
                        dy: DY,
                        'text-anchor': ANCHOR,
                        transform: `rotate(${rotation} ${X} ${Y})`
                    });

                    let temp_child = temp_label.node().childNodes;
                    const boundWithTspan = function (temp_label) {
                        let htmlContent = `<tspan>${temp_label.node().innerHTML}</tspan>`;
                        temp_label.node().innerHTML = htmlContent;
                        return "html";
                    }
                    
                    contentType = (temp_child.length) ?
                        (temp_child[0].childNodes.length) ?
                            (temp_child[0].childNodes[0].childNodes.length) ?
                                contentType : boundWithTspan(temp_label) : "text" : contentType;

                    if ((isFixedTopPosition || isStackedLabels) && display !== false) {
                        var maxWidth = perToPx(defined(dataLabels.maxWidth) ? dataLabels.maxWidth : '30%', chartObj.plotarea.width);

                        if (contentType === 'html') {
                            translateTspan(temp_label, X);
                            var bb = bbox(temp_label.node()),
                                edge = {
                                    left: bb.x,
                                    top: bb.y,
                                    right: bb.x + maxWidth,
                                    bottom: bb.y + bb.height
                                },
                                actualWidth = bb.width;
                            var tspanInfo = getTspanBound(temp_label, fontSize);
                            trimAllTspan(temp_label, chartObj, [0, 0], 0, edge, tspanInfo);
                            text = temp_label.html();
                        } else {
                            var bb = bbox(temp_label.node()),
                                edge = {
                                    left: bb.x,
                                    top: bb.y,
                                    right: bb.x + maxWidth,
                                    bottom: bb.y + bb.height
                                };
                            callTrimText(temp_label, chartObj, [0, 0], 0, edge);
                            text = temp_label.text();
                        }
                    }

                    var xDiff = 0,
                        yDiff = 0,
                        adjustX = 0,
                        adjustY = 0;
                    HOOKPOSITION = labelPos === MIDDLE_ALIGN ? NULL : HOOKPOSITION; // There is no hook for middle datalabels
                    if (axisrotated) {
                        adjustY = (textDirection === HORIZONTAL_ALIGN ? widthToBeAdded : heightToBeAdded) / 2;
                    } else {
                        adjustX = (textDirection === HORIZONTAL_ALIGN ? widthToBeAdded : heightToBeAdded) / 2;
                    }

                    if (!clip) {
                        if (isMultiLine) {
                            translateTspan(temp_label, X);
                        }
                        var adjust = boundBoxPadding + boundBoxStrokeWidth / 2,
                            scrolledX = 0, //chartObj.chartRendered ? (chartObj.scrolledX || 0) : 0,
                            scrolledY = 0; //chartObj.chartRendered ? (chartObj.scrolledY || 0) : 0;

                        var b = temp_label.node().getBoundingClientRect(),
                            leftPos = b.left - rectBoundLeft - adjust - adjustX - scrolledX,
                            rightPos = leftPos + b.width + 2 * adjust + 2 * adjustX,
                            topPos = b.top - rectBoundTop - adjust - adjustY - scrolledY,
                            bottomPos = topPos + b.height + 2 * adjust + 2 * adjustY;

                        if (renderWithinPlot) {
                            var newX = X + (leftPos < left ? left - leftPos : rightPos > right ? -(rightPos - right) : 0);
                            if (newX != X && !(isFixedTopPosition && axisrotated)) {
                                xDiff = newX - X;
                                X = newX;
                            }
                            var newY = Y + (topPos < top ? top - topPos : bottomPos > bottom ? -(bottomPos - bottom) : 0);
                            if (newY != Y && !(isFixedTopPosition && !axisrotated)) {
                                yDiff = newY - Y;
                                Y = newY;
                            }
                        }
                    }

                    /**
                     * Overlapping handle
                     */

                    var isWithinClrRange = isWithinColorRange(dataset.getClr(d), chartObj);
                    display = display && (isStackedLabels || isWithinClrRange);
                    if (handleOverlapping) {
                        if (stacked && !isStackedLabels) {
                            temp_label //Test_828.json
                                .attrs({
                                    x: X,
                                    y: Y
                                });
                            if (isMultiLine) {
                                translateTspan(temp_label, X);
                            }
                            boundBox = bbox(temp_label.node());
                            var x = getParamValue('x', d, i, j, commonOption) || 0,
                                y = getParamValue('y', d, i, j, commonOption) || 0,
                                height = getParamValue('height', d, i, j, commonOption) || 0,
                                width = getParamValue('width', d, i, j, commonOption) || 0;

                            if (yaxisReversed) {
                                axisrotated ? (x -= width) : (y -= height);
                            }

                            var labelBound = {
                                x1: boundBox.x - boundBoxPadding,
                                y1: boundBox.y - boundBoxPadding,
                                x2: boundBox.x + boundBox.width + boundBoxPadding,
                                y2: boundBox.y + boundBox.height + boundBoxPadding
                            },
                                barBound = {
                                    x1: x,
                                    y1: y,
                                    x2: x + width,
                                    y2: y + height
                                };

                            var within = isLabelWithinBound(labelBound, barBound);
                            if (!within) {
                                display = false;
                            } else if (cType_Obj.name === BULLET_CHARTNAME || isComboChart) {   //ZC4386
                                var b = temp_label.node().getBoundingClientRect(),
                                    bx = b.left - rectBoundLeft,
                                    by = b.top - rectBoundTop;
                                var obj = {
                                    id: `lbl${rendererIndex}_${i}_${j}_${index}_${yrangePos}`,
                                    translateFunction: updateHandler,
                                    isStackedLabels: isStackedLabels,
                                    x1: bx - boundBoxPadding + xDiff - adjustX,
                                    y1: by - boundBoxPadding + yDiff - adjustY,
                                    x2:
                                        bx +
                                        b.width +
                                        boundBoxPadding +
                                        xDiff +
                                        (!rotation ? 0 : hookHeight + 2 * boundBoxStrokeWidth) +
                                        adjustX,
                                    y2:
                                        by +
                                        b.height +
                                        boundBoxPadding +
                                        yDiff +
                                        (rotation ? 0 : hookHeight + 2 * boundBoxStrokeWidth) +
                                        adjustY,
                                    overlap: isFixedTopPosition && textOverlap !== 'hidden' && textOverlap !== NONE ? 'hidden' : textOverlap
                                };
                                labelArray.push(obj);
                            }
                        } else {
                            if (!clip || isStackedLabels) {
                                if (isMultiLine) {
                                    translateTspan(temp_label, X);
                                }
                                var b = temp_label.node().getBoundingClientRect();
                                var bx = b.left - rectBoundLeft,
                                    by = b.top - rectBoundTop;
                                var obj = {
                                    id: `lbl${rendererIndex}_${i}_${j}_${index}_${yrangePos}`,
                                    translateFunction: updateHandler,
                                    isStackedLabels: isStackedLabels,
                                    x1: bx - boundBoxPadding + xDiff - adjustX,
                                    y1: by - boundBoxPadding + yDiff - adjustY,
                                    x2:
                                        bx +
                                        b.width +
                                        boundBoxPadding +
                                        xDiff +
                                        (!rotation ? 0 : hookHeight + 2 * boundBoxStrokeWidth) +
                                        adjustX,
                                    y2:
                                        by +
                                        b.height +
                                        boundBoxPadding +
                                        yDiff +
                                        (rotation ? 0 : hookHeight + 2 * boundBoxStrokeWidth) +
                                        adjustY,
                                    overlap: isFixedTopPosition && textOverlap !== 'hidden' && textOverlap !== NONE ? 'hidden' : textOverlap
                                };
                                labelArray.push(obj);
                            } else {
                                if (textDirection != HORIZONTAL_ALIGN) {
                                    var w = getParamValue(axisrotated ? 'width' : 'height', d, i, j, commonOption) || 0;
                                    width =
                                        labelPos == TOP_ALIGN
                                            ? boundBox.width
                                            : w - (labelPos == BOTTOM_ALIGN ? padding + boundBoxPadding : 2 * boundBoxPadding);
                                } else {
                                    width = barW - 2 * boundBoxPadding;
                                }

                                boundBox.width = mathMin(width, boundBox.width);
                            }
                        }
                    }

                    var coordinate = {
                        X: X,
                        Y: Y,
                        DY: DY,
                        ANCHOR: ANCHOR,
                        text: text,
                        display: display,
                        height: boundBox.height + heightToBeAdded,
                        width: boundBox.width + widthToBeAdded,
                        hookPosition: HOOKPOSITION,
                        adjustVal: textDirection === HORIZONTAL_ALIGN ? (axisrotated ? yDiff : xDiff) : NULL
                    };

                    if (actualDY !== undefined) {
                        coordinate.actualDY = actualDY;
                    }

                    if (isCanvasMode && contentType === 'html') {
                        coordinate.tspanInfo = getTspanBound(temp_label, fontSize);
                    }

                    commonProperties.push(coordinate);
                });

                // add to drawing queue
                commonProperties.dataUpdate = dataUpdate; //Each subseries may have different dataUpdates
                var q = queue(commonProperties, index, i, yrangePos, isStackedLabels ? STACKLABELS_LABELTYPE : NULL, rendererIndex);
                q.properties = {
                    type: BAR_CHARTNAME,
                    dataLabels: dataLabels,
                    rotation: rotation,
                    color: rp.color,
                    dataUpdate: dataUpdate,
                    showlabel: showlabel,
                    multiColoring: multiColoring,
                    textDirection: textDirection,
                    rangeChart: hasRangePos,
                    yrangePos: yrangePos,
                    rp: rp,
                    contentType: contentType,
                    isMultiLine: isMultiLine,
                    chartType: cType_Obj.value
                };
            });

            function getFixedTopPosition() {
                if (datatypeY === ORDINAL_DATATYPE) {
                    var domain = yscale.domain();
                    return yscale(domain[domain.length - 1]) + yscale.bandwidth() / 2;
                } else {
                    var rangeCorrections_datalabels = chartObj.cache.correction || {},
                        maxDatalabelSize = [0],
                        operation = axisrotated ? (yaxisReversed ? 1 : -1) : yaxisReversed ? -1 : 1;
                    for (var yaxis in rangeCorrections_datalabels) {
                        if (yaxis.indexOf('label-y') > -1) {
                            maxDatalabelSize.push(d3_max(rangeCorrections_datalabels[yaxis]));
                        }
                    }

                    return yscale(yscale.domain()[1]) + d3_max(maxDatalabelSize) * operation;
                }
            }
        });
    }

    labels.renderer = function (_) {
        if (!arguments.length) {
            return renderer;
        }
        renderer = _;
        return labels;
    };
    labels.args = function (_) {
        if (!arguments.length) {
            return args;
        }
        args = _;
        return labels;
    };

    function getUniqueDataByX(data) {
        var xCategories = d3_set();
        return data.filter(function (d) {
            var xVal = d[0];
            if (xCategories.has(xVal)) {
                return false;
            } else {
                xCategories.add(xVal);
                return true;
            }
        });
    }

    function getVisibleData(data, xCategories) {
        return data.filter(function (d) {
            var x = d[0];
            return xCategories.indexOf(x) > -1;
        });
    }

    return labels;
}

LabelRegistry.add('bar', barLabel);
export var _barDatalabel = barLabel;
