// @zohocharts/zc/src/label/boundBoxUtils.js
//$Id$
//boundBoxUtils.js
import { getRawColor } from '../utils/colorutils';
import { mergeJSON, validateNullVal, getValByPriority, defined, isUndefined } from '../utils/zcutils';
import { mathMax } from '../core/zc-constants';
import { getIndexForItemColor } from '../utils/colorutils';

export function getBoundBoxColor(actualColor, properties, chartObj, data, index) {
    var data_length = data.length;
    return (
        actualColor ||
        function(d, meta) {
            var clrIndex = getIndexForItemColor(meta.itemIndex, index, meta, chartObj);
            return getRawColor(chartObj, clrIndex, d && d.data ? d.data : d, index);
        }
    );
}

export function getMergedBoundBoxOptions(dataLabels, chartObj) {
    return mergeJSON(chartObj.systemConf.chart.plot.datalabels.boundBox, dataLabels.boundBox, true);
}

export function isBoundBoxEnabled(dataLabels, chartObj) {
    return getValByPriority([dataLabels.boundBox || {}, chartObj.systemConf.chart.plot.datalabels.boundBox || {}], 'show') === true;
}

export function getBoundBoxPadding(dataLabels, chartObj) {
    return mathMax(
        validateNullVal(
            getValByPriority([dataLabels.boundBox || {}, chartObj.systemConf.chart.plot.datalabels.boundBox || {}], 'padding'),
            2
        ),
        0
    );
}

export function getBoundBoxStrokeWidth(dataLabels, chartObj) {
    return validateNullVal(
        getValByPriority([dataLabels.boundBox || {}, chartObj.systemConf.chart.plot.datalabels.boundBox || {}], 'strokeWidth'),
        1
    );
}

export function getSizeToBeAdded(defaultWidth, padding, radius, hookWidth) {
    var datalabels_min_width = 2 * (radius || 0) + hookWidth;
    return mathMax(datalabels_min_width - defaultWidth - 2 * padding, 0);
}

/*
 * Returns the options for boundbox hook
 * @param {Object} hookOptions
 * @return {Object} obj which contains enabled, width, height
 */
export function getBoundBoxHookConf(hookOptions) {
    hookOptions = hookOptions || {};
    var hookEnabled = hookOptions.enabled === true;
    var obj = {};

    obj.enabled = hookEnabled;
    obj.width = hookEnabled ? (defined(hookOptions.width) ? +hookOptions.width : 10) : 0;
    obj.height = hookEnabled ? (defined(hookOptions.height) ? +hookOptions.height : 4) : 0;
    return obj;
}

/**
 * Visibility property of bound box
 * @param  {Object}  d
 * @return {Boolean}
 */
export function hasBoundBox(d) {
    return !(d.display === false || isUndefined(d.text) || d.text === '');
}

export function getMetaConfigForBoundbox(itemIndex, seriesIndex, properties) {
    return {
        seriesIndex: seriesIndex,
        itemIndex: itemIndex,
        multiColoring: properties.multiColoring,
        reverseColor: properties.reverseColor,
        dataCount: properties.dataUpdate.length
    };
}
