// @zohocharts/zc/src/label/datalabelsUtils.js
//$Id$
//datalabelsUtils.js
import { Registry } from '../core/zc-registry';
import {
    mergeJSON,
    isArray,
    defined,
    invokeFunction,
    getPropVal,
    pick,
    getValByPriority,
    isObjectPropDefined,
    bbox,
    getContrastColor
} from '../utils/zcutils';
import { isBarChartType } from '../events/events-utils';
import { UNDEFINED, TRANSPARENT } from '../core/zc-constants';
import { chartcategory, PERCENT_PATTERN } from '../core/main-zc-init';
import { datacount_based_intelligence } from '../utils/intelligence-utils';
import { DATALABELS_LABELTYPE } from './labeltype-constants';
import { PERCENT_SHOWAS, SPLITPERCENT_SHOWAS, MAXPERCENT_SHOWAS } from './showas-constants';
import { BOTTOM_ALIGN } from '../core/align-constants';
import { $Browser } from '../core/browser-detect';
import { trimText, applyFont } from '../utils/zcsvgutils';
import { getMergedBoundBoxOptions, isBoundBoxEnabled } from './boundBoxUtils';
import { $ZCG } from '../core/zcglobal';
import { getChartTypeByIndex } from '../core/charttype-helpers';
import { SANKEY_CHARTNAME } from '../core/chartname-constants';

export function getMergedDatalabelOptions(commonPlotOptions, plotoptions, chartObj, index, paneid) {
    var options =
        plotoptions && plotoptions.datalabels
            ? mergeJSON(commonPlotOptions.datalabels, plotoptions.datalabels, true)
            : commonPlotOptions.datalabels;
    if (chartObj && chartObj.viewFlow && chartObj.viewFlow.marks._default.label.length) {
        let marks_plot = chartObj.viewFlow.marks._default.label[index],
            markLabelOptions;
        markLabelOptions = marks_plot.spec;
        options = markLabelOptions ? mergeJSON(markLabelOptions, options, true) : options;
    }
    if (chartObj) {
        overrideDatalabelShowAs(options, chartObj, index, paneid);
    }

    return options;
}

export function applyDatalabelFonts(labelseries, dataLabels, chartObj, seriesColor, inner, seriesIndex, isStackLabel) {
    var systemconf = chartObj.systemConf;

    if (isArray(seriesColor) && defined(seriesIndex)) {
        seriesColor = seriesColor[seriesIndex];
    }

    let _fontColor = getLabelFontColor(systemconf, seriesColor, dataLabels, chartObj, inner, seriesIndex, isStackLabel, 'fontColor'),
        _strokeColor = getLabelFontColor(systemconf, seriesColor, dataLabels, chartObj, inner, seriesIndex, isStackLabel, 'strokeColor');

    return applyFont(labelseries, [
        { fontColor: _fontColor, strokeColor: _strokeColor }, // Series color
        dataLabels,
        systemconf.chart.plot.datalabels,
        systemconf.chart,
        systemconf.canvas,
        $ZCG
    ]);
}
Registry.setComponent('dataLabel.applyFont', applyDatalabelFonts);

export function applyTextShadow4Firefox(stroke, strokeWidth = 1, textShadow) {
    if (!($Browser.FIREFOX && textShadow && strokeWidth)) {
        return;
    }

    // ZC2858
    if (stroke === TRANSPARENT) {
        return { stroke: null };
    }

    return {
        stroke: null,
        'text-shadow': strokeInShadow(stroke, strokeWidth, textShadow)
    };
}

function strokeInShadow(color, size, shadow) {
    size = parseInt(size / 2);
    let p = shadow.split(/ (?![^(]*\))/),
        ts = `${parseFloat(p[0]) + size}px ${parseFloat(p[1]) + size}px ${p[2]} ${p[3] || 'black'}`;

    return `${-size}px ${-size}px 0 ${color}, ${size}px ${-size}px 0 ${color}, ${-size}px ${size}px 0 ${color}, ${size}px ${size}px 0 ${color}, ${ts}`;
}

export function getLabelFontColor(
    systemconf,
    seriesColor,
    config,
    chartObj,
    inner,
    seriesIndex,
    isStackLabel,
    styleName = 'fontColor',
    paneid
) {
    /**
     * @param inner is boolean, means function being call from inner-label
     */

    var fontColor = config[styleName];
    if (typeof fontColor === 'function') {
        var userColor = invokeFunction(fontColor, seriesColor, chartObj);
        if (defined(userColor)) {
            return userColor;
        }
    }

    var globalConfig = getPropVal(chartObj.systemConf, 'chart.plot.datalabels') || {},
        mapLabelOptions = getPropVal(chartObj.systemConf, 'map.labels') || {},
        dataLabelColor = chartObj.isGeoMap ? pick(globalConfig[styleName], mapLabelOptions[styleName]) : globalConfig[styleName],
        chart = getChartTypeByIndex(chartObj, paneid, seriesIndex || 0),
        chartNumber = chart.value,
        imagePalletUsed = defined(systemconf.legend.imagePallete),
        color,
        isStackedBar = isBarChartType(chartNumber) && chartObj.dataObject.isstacked(chartNumber) && isStackLabel !== true,
        sankeyInsideLabel = chart.name === SANKEY_CHARTNAME && config.labelPos === 'inside',
        boundBoxColor;

    if (isBoundBoxEnabled(config, chartObj)) {
        //get the boundBox color
        var boundboxOptions = getMergedBoundBoxOptions(config, chartObj);
        boundBoxColor = boundboxOptions.fillColor || seriesColor;
    } else if (sankeyInsideLabel) {
        let nodeOptions = systemconf.chart.plot.plotoptions.sankey.node;
        seriesColor = pick(nodeOptions.fillColor, seriesColor);
    }

    // for constant color
    if (styleName === 'fontColor' && fontColor === UNDEFINED && dataLabelColor === UNDEFINED) {
        if (boundBoxColor && boundBoxColor !== TRANSPARENT) {
            return getContrastColor(boundBoxColor); //#ZC1232
        } else if (inner && imagePalletUsed) {
            return '#444444'; // always returns default color
        } else if (
            (inner && chartcategory.innerLabels.indexOf(chartNumber) > -1) ||
            isStackedBar ||
            sankeyInsideLabel ||
            chartcategory.datalabelsInside.indexOf(chartNumber) > -1
        ) {
            return getContrastColor(seriesColor);
        }
    }

    color = !imagePalletUsed && ((!defined(fontColor) && dataLabelColor === null) || fontColor === null) ? seriesColor : fontColor;
    return color || pick(dataLabelColor, systemconf.chart[styleName], systemconf.canvas[styleName], $ZCG[styleName]);
}

export function isLabelWithinBound(labelbound, plotbound) {
    return !(labelbound.x1 < plotbound.x1 || labelbound.y1 < plotbound.y1 || labelbound.x2 > plotbound.x2 || labelbound.y2 > plotbound.y2);
}

export function isDatalabelEnabled(dataLabels, chartObj) {
    if (datacount_based_intelligence(chartObj, DATALABELS_LABELTYPE)) {
        let dataSplicer = chartObj.dataSplicer,
            metadata_axes = dataSplicer.getEncode(),
            labelMeta = metadata_axes.label;
        return (
            (labelMeta && defined(labelMeta[0])) ||
            getValByPriority([dataLabels, chartObj.systemConf.chart.plot.datalabels], 'show') === true
        );
    }
    return false;
}
Registry.setComponent('dataLabel.isDatalabelEnabled', isDatalabelEnabled);

export function isStacklabelEnabled(stackLabels, chartObj) {
    return (
        chartObj.dataObject.noofBarEnabledSeries &&
        chartObj.dataObject.isstacked() &&
        !chartObj.dataObject.isStackedPercentChart &&
        stackLabels &&
        stackLabels.show
    );
}

export function isShowAsPercentage(chartName, chartObj) {
    var datalabelOptions = chartObj.systemConf.chart.plot.plotoptions[chartName].datalabels || {},
        userDatalabelsOptions = isObjectPropDefined(chartObj.userdata, `chart.plot.plotoptions.${chartName}.datalabels`, true) || {},
        showAs = userDatalabelsOptions.showAs || datalabelOptions.showAs;
    return showAs === PERCENT_SHOWAS || showAs === SPLITPERCENT_SHOWAS || showAs === MAXPERCENT_SHOWAS || PERCENT_PATTERN.test(showAs);
}

export function callTrimText(labelText, chartObj, tranform, fontSize, edge, xy) {
    var dx = +tranform[0],
        dy = +tranform[1],
        plotarea = chartObj.plotarea,
        margin = {
            left: -fontSize,
            right: -fontSize,
            top: 0,
            bottom: 0
        },
        pattern = '..',
        orient = BOTTOM_ALIGN;

    edge = edge || {
        left: -dx,
        top: -dy,
        right: plotarea.width - dx,
        bottom: plotarea.height - dy
    };
    labelText.each(function() {
        var bb = bbox(this);
        var textBound = {
            left: bb.x + (xy ? xy[0] : 0),
            top: bb.y + (xy ? xy[1] : 0),
            width: bb.width,
            height: bb.height
        };
        trimText(this, this.textContent, textBound, edge, orient, pattern, margin);
    });
}

function overrideDatalabelShowAs(options, chartObj, index, paneid) {
    /* If defaulttheme.showAs ='y' && userdata.showAs = ['<tspan>{{val(0)}}</tspan>', '<tspan style="fill:red">{{val(1)}}</tspan>'], merged sysconf will have "y" only.
     * The following will update the userdata's showAs if presents.
     */
    var cType_Obj = getChartTypeByIndex(chartObj, paneid, index),
        chartName = cType_Obj.name;
    var userDatalabels = isObjectPropDefined(chartObj.userdata, `chart.plot.plotoptions.${chartName}.datalabels`, true) || {},
        userSeriesSpecificDatalabels =
            isObjectPropDefined(chartObj.seriesdata[index] || chartObj.seriesdata[0], 'plotoptions.datalabels', true) || {},
        userInnerLabel,
        userParentLabel,
        userStackLabel,
        systemConfInnerLabel,
        systemConfParentLabel,
        systemConfStackLabel;

    if (userSeriesSpecificDatalabels.showAs || userDatalabels.showAs) {
        //#ZC1663
        options.showAs = userSeriesSpecificDatalabels.showAs || userDatalabels.showAs || options.showAs;
    }

    if ((userInnerLabel = userDatalabels.innerLabel)) {
        systemConfInnerLabel = options.innerLabel;
        systemConfInnerLabel.showAs = userInnerLabel.showAs || systemConfInnerLabel.showAs;

        if ((userParentLabel = userInnerLabel.parentLabel)) {
            systemConfParentLabel = options.innerLabel.parentLabel;
            systemConfParentLabel.showAs = userParentLabel.showAs || systemConfParentLabel.showAs;
        }
    } else if ((userStackLabel = userDatalabels.stackLabels)) {
        systemConfStackLabel = options.stackLabels;
        systemConfStackLabel.showAs = userStackLabel.showAs || systemConfStackLabel.showAs;
    }
}

export function getLabelHolder(chartObj) {
    return chartObj.staticPaneSVG.select('.labelholder');
}

export function getLabelArray(rp, paneid, subIndex) {
    let { labelArray } = rp,
        key = `${paneid || 0}`;

    if (key.indexOf('_') === -1) {
        key = '0';
    } else if (defined(subIndex)) {
        key += `(${subIndex})`;
    }

    labelArray[key] = labelArray[key] || [];
    return labelArray[key];
}
