// @zohocharts/zc/src/label/dialdatalabels.js
//$Id$
//dialdatalabels.js
import { charttype } from '../core/main-zc-init';
import { LabelRegistry } from './label-registry';
import { getMergedDatalabelOptions, isDatalabelEnabled, applyDatalabelFonts } from './datalabelsUtils';
import { dataLabelTextContent } from './datalabel-content';
import { isBoundBoxEnabled, getBoundBoxStrokeWidth, getBoundBoxPadding } from './boundBoxUtils';
import { DATALABELS_LABELTYPE } from './labeltype-constants';
import { appendEle } from '../utils/zcsvgutils';
import { bbox, defined, invokeFunction } from '../utils/zcutils';
import { translateTspan } from './tspan-utils';
import { DIAL_CHARTNAME } from '../core/chartname-constants';

function dialLabel() {
    var renderer, args;

    function labels(selection, labelHandler = {}, container) {
        selection.each(function() {
            var currentdata = args[0],
                index = args[1],
                rp = renderer.commonRendererProp().params,
                chartObj = rp.chartObj,
                commonPlotOptions = rp.commonPlotOptions,
                color = rp.color,
                { rendererIndex = index } = rp,
                paneid = labelHandler.paneid,
                plotoptions = currentdata.plotoptions;
            var plotarea_approx = chartObj.plotarea_approx,
                yscaleDomain = chartObj.axes.y.groups[0].scale.domain();

            var dataLabels = getMergedDatalabelOptions(commonPlotOptions, plotoptions, chartObj, index);
            var showlabel = isDatalabelEnabled(dataLabels, chartObj),
                GLT = dataLabelTextContent(
                    chartObj,
                    rp,
                    dataLabels,
                    DATALABELS_LABELTYPE,
                    yscaleDomain[1] - yscaleDomain[0],
                    0,
                    undefined,
                    paneid
                ),
                contentType = GLT.contentType || 'text',
                isMultiLine = contentType === 'html' && GLT.multiLine;

            if (!showlabel) {
                return;
            }

            var boundBoxPadding = isBoundBoxEnabled(dataLabels, chartObj)
                ? getBoundBoxPadding(dataLabels, chartObj) + getBoundBoxStrokeWidth(dataLabels, chartObj)
                : 0;

            var temp_label = appendEle(chartObj.staticPaneSVG.select('.labelholder'), 'text', [1], '.', 'class', 'temp_label');
            applyDatalabelFonts(temp_label, dataLabels, chartObj, color[0]);
            temp_label.style('opacity', 0);

            var dataUpdate = [currentdata[0].data[0][0]],
                commonProperties = [],
                i = 0; //subseries index

            dataUpdate.forEach(function(d, j) {
                var ANCHOR = 'middle',
                    DY = isMultiLine ? '0.9em' : '-.3em',
                    text = GLT(d, j),
                    display = true;

                temp_label[contentType](text).attrs({
                    x: X,
                    dy: DY,
                    'text-anchor': ANCHOR
                });

                var bb = bbox(temp_label.node());
                var X = defined(dataLabels.x) ? invokeFunction(dataLabels.x, d, chartObj, bb) : plotarea_approx.x,
                    Y = defined(dataLabels.y)
                        ? invokeFunction(dataLabels.y, d, chartObj, bb)
                        : plotarea_approx.y +
                          (commonPlotOptions.needle.show
                              ? bb.height + (+dataLabels.padding || 0) + (+commonPlotOptions.needle.width / 2 || 7) + boundBoxPadding
                              : bb.height / 2);

                if (isMultiLine) {
                    if (!defined(dataLabels.y)) {
                        Y -= bb.height;
                    }
                    translateTspan(temp_label, X);
                    bb = bbox(temp_label.node());
                }

                var coordinate = {
                    X: X,
                    Y: Y,
                    DY: DY,
                    ANCHOR: ANCHOR,
                    text: text,
                    display: display,
                    ID: `lbl${index}_${i}_${j}`,
                    width: bb.width,
                    height: bb.height
                };
                commonProperties.push(coordinate);
            });

            if (container) {
                var conf = commonProperties[0];
                var labelText = appendEle(container, 'text', showlabel ? dataUpdate : []);
                applyDatalabelFonts(labelText, dataLabels, chartObj, color[0]);
                labelText[contentType](conf.text).attrs({
                    'text-anchor': conf.ANCHOR,
                    dy: conf.DY,
                    transform: `translate(${conf.X},${conf.Y})`
                });

                if (isMultiLine) {
                    translateTspan(temp_label, conf.X);
                }
            } else {
                var queue = labelHandler.add;
                // add to drawing queue
                var q = queue(commonProperties, index, i, null, null, rendererIndex, paneid);
                q.properties = {
                    type: DIAL_CHARTNAME,
                    dataLabels: dataLabels,
                    dataUpdate: dataUpdate,
                    showlabel: showlabel,
                    rp: rp,
                    contentType: contentType,
                    isMultiLine: isMultiLine,
                    chartType: charttype.dial
                };
            }
        });
    }

    labels.renderer = function(_) {
        if (!arguments.length) {
            return renderer;
        }
        renderer = _;
        return labels;
    };
    labels.args = function(_) {
        if (!arguments.length) {
            return args;
        }
        args = _;
        return labels;
    };
    return labels;
}

LabelRegistry.add('dial', dialLabel);
export var _dialDatalabel = dialLabel;
