// @zohocharts/zc/src/label/funneldatalabels.js
//$Id$
//funneldatalabels.js
import { LabelRegistry } from './label-registry';
import { PYRAMID_CHARTNAME, PIE_CHARTNAME, FUNNEL_CHARTNAME } from '../core/chartname-constants';
import { d3_line } from '../core/d3-constants';
import { getMergedDatalabelOptions, isDatalabelEnabled, applyDatalabelFonts, getLabelArray } from './datalabelsUtils';
import {
    isBoundBoxEnabled,
    getMergedBoundBoxOptions,
    getBoundBoxHookConf,
    getBoundBoxPadding,
    getBoundBoxStrokeWidth
} from './boundBoxUtils';
import { getFontSizeByHierarchy } from '../utils/zcdatautils';
import { dataLabelTextContent } from './datalabel-content';
import { DATALABELS_LABELTYPE } from './labeltype-constants';
import { appendEle } from '../utils/zcsvgutils';
import { getSizeToBeAdded } from './boundBoxUtils';
import { bbox, defined, isUndefined, pick, isObjectPropDefined, invokeFunction } from '../utils/zcutils';
import { LEFT_ALIGN, RIGHT_ALIGN } from '../core/align-constants';
import { isWithinColorRange, isSeriesFiltered, hasColorColumnInNonAxis } from '../utils/colorutils';
import { mathMax } from '../core/zc-constants';
import { translateTspan } from './tspan-utils';
import { getSvgTransform } from '../utils/svg-transforms';
import { getChartTypeByIndex } from '../core/charttype-helpers';

function funnelLabel() {
    var MARGIN = 5,
        d3Line = d3_line(),
        renderer,
        args;

    function labels(selection, labelHandler) {
        selection.each(function() {
            var currentdata = args[0],
                index = args[1],
                data = args[2],
                noofNotNullData = args[3],
                arcHeight = args[4],
                totalEngagement = args[5],
                cfg = args[6],
                labelPosition = args[7];
            var paneid = labelHandler.paneid,
                chartWidth = cfg.width,
                chartHeight = cfg.height,
                rp = renderer.commonRendererProp().params,
                chartObj = rp.chartObj,
                systemconf = chartObj.systemConf,
                commonPlotOptions = rp.commonPlotOptions,
                dataset = rp.dataset,
                yaxiscolumnorder = rp.yaxiscolumnorder,
                labelArray = getLabelArray(rp, paneid, args[8]),
                { rendererIndex = index } = rp,
                { name: chartName, value: cType } = getChartTypeByIndex(chartObj, paneid, index),
                chartTop = (renderer.plotarea && renderer.plotarea() && renderer.plotarea().top) || 0,
                dataLabels;
            var translation = `translate(${args[9]})`,
                isPyramid = chartName === PYRAMID_CHARTNAME;

            currentdata = currentdata[index];
            var plotoptions = currentdata.plotoptions,
                boundBoxPaddingAlone = 0,
                boundBoxPadding = 0,
                hookWidth = 0,
                boundBoxRadius = 0;
            dataLabels = getMergedDatalabelOptions(commonPlotOptions, plotoptions, chartObj, index);

            if (isBoundBoxEnabled(dataLabels, chartObj)) {
                var boundBoxOptions = getMergedBoundBoxOptions(dataLabels, chartObj);
                var hookOptions = getBoundBoxHookConf(boundBoxOptions.hook);
                boundBoxPaddingAlone = getBoundBoxPadding(dataLabels, chartObj);
                boundBoxPadding = boundBoxPaddingAlone + getBoundBoxStrokeWidth(dataLabels, chartObj) + hookOptions.height;
                hookWidth = hookOptions.width;
                boundBoxRadius = boundBoxOptions.radius;
            }

            var padding = cfg.datalabelPadding || +dataLabels.padding || 0,
                type = dataLabels.type,
                handleOverlapping = systemconf.chart.plot.datalabels.handleOverlapping === true,
                txtHgt = getFontSizeByHierarchy(dataLabels, chartObj),
                GLT = dataLabelTextContent(chartObj, rp, dataLabels, DATALABELS_LABELTYPE, totalEngagement, index, undefined, paneid),
                contentType = GLT.contentType || 'text',
                isHTMLContent = contentType === 'html',
                isMultiLine = isHTMLContent && GLT.multiLine,
                showlabel = isDatalabelEnabled(dataLabels, chartObj),
                relativeX = dataLabels.relativeX === true,
                useBBox = isObjectPropDefined(dataLabels, 'preCalculation.useBBox', true);

            if (!showlabel) {
                return;
            }

            showlabel = showlabel ? type : 'hidden';
            var data_length = data.length,
                PathGroupPosition = getSvgTransform(chartObj.svg.selectAll(`g.${chartName}group`)).translate,
                idx = 0;

            var temp_label = appendEle(chartObj.staticPaneSVG.select('.labelholder'), 'text', [1], '.', 'class', 'temp_label'),
                node = temp_label.node();
            applyDatalabelFonts(temp_label, dataLabels, chartObj, rp.color);
            temp_label.attr('transform', translation).style('opacity', 0);
            var queue = labelHandler.add;

            (hasColorColumnInNonAxis(chartObj) ? [currentdata] : currentdata.data).forEach(function(d, i) {
                var dataUpdate = labelPosition,
                    commonProperties = [],
                    subSeriesIndex = pick(args[8], i),
                    overlap = {},
                    objs = [];

                dataUpdate.forEach(function(d, j) {
                    var display = !isdisabled(j),
                        isLeft = isLeftHalf(idx),
                        ANCHOR = isLeft ? 'end' : null,
                        DY = isMultiLine && isUndefined(dataLabels.y) ? '0.9em' : '0.3em',
                        dt = (d.data = data[data_length - 1 - j]),
                        text = GLT(dt, data_length - 1 - j),
                        //
                        xF = pick(dataLabels.x, (d, chartObj, bb) => {
                            let _ = getEndX(d, idx, overlap) + (isLeft ? -boundBoxPadding : boundBoxPadding);
                            if (display && isMultiLine && !useBBox) {
                                _ += (bb.width / 2) * (isLeft ? -1 : 1);
                            }
                            return _;
                        }),
                        yF = pick(dataLabels.y, (d, chartObj, bb) => {
                            let _ = isPyramid && isLeft ? d.leftY : mathMax(d.y, txtHgt - arcHeight - d.y);
                            if (display && isMultiLine) {
                                _ -= bb.height / 2;
                            }
                            return _;
                        }),
                        heightToBeAdded = 0,
                        bb;

                    temp_label[contentType](text).attrs({
                        'text-anchor': ANCHOR,
                        dy: DY
                    });

                    if (display && isMultiLine) {
                        translateTspan(temp_label, 0);
                    }

                    bb = bbox(node);
                    let X = invokeFunction(xF, d, chartObj, bb),
                        Y = invokeFunction(yF, d, chartObj, bb);

                    if (display && isMultiLine) {
                        translateTspan(temp_label, X);
                    }

                    temp_label.attrs({
                        x: X,
                        y: Y
                    });

                    if (handleOverlapping) {
                        if (display) {
                            bb = bbox(node);
                            heightToBeAdded = getSizeToBeAdded(bb.height, boundBoxPaddingAlone, boundBoxRadius, hookWidth);
                            var DX = isLeft ? -(0 + padding + boundBoxPadding) : chartWidth + padding + boundBoxPadding;
                            if (isMultiLine && !useBBox) {
                                DX += (bb.width / 2) * (isLeft ? -1 : 1);
                            }

                            objs.push({
                                id: `lbl${index}_${subSeriesIndex}_${j}`,
                                x1: bb.x - boundBoxPadding,
                                x2: bb.x + bb.width + boundBoxPadding,
                                y1: bb.y - MARGIN - boundBoxPadding - heightToBeAdded / 2,
                                isLeft: isLeft,
                                y2: bb.y + bb.height + MARGIN + boundBoxPadding + heightToBeAdded / 2,
                                DX: DX - X
                            });
                        }
                    } else {
                        bb = bbox(node);
                        heightToBeAdded = getSizeToBeAdded(bb.height, boundBoxPaddingAlone, boundBoxRadius, hookWidth);
                    }

                    var coordinate = {
                        X,
                        Y,
                        ANCHOR,
                        DY,
                        text,
                        display,
                        ID: `lbl${index}_${subSeriesIndex}_${j}`,
                        width: display ? bb.width : 0,
                        height: display ? bb.height + heightToBeAdded : 0,
                        hookPosition: isLeft ? RIGHT_ALIGN : LEFT_ALIGN,
                        XY: [X, Y]
                    };

                    if (isHTMLContent) {
                        coordinate.isLeft = isLeft;
                    }

                    commonProperties.push(coordinate);

                    idx += display ? 1 : 0;
                });

                idx = -1;
                commonProperties.id = `${paneid}(${index}_${subSeriesIndex})`;
                commonProperties.translation = translation;
                commonProperties.dataUpdate = dataUpdate;
                commonProperties.overlap = overlap;
                commonProperties.connectorPath = function(d, i) {
                    idx += isdisabled(i) ? 0 : 1;
                    var coordinate = commonProperties[i],
                        y = coordinate.Y + (isMultiLine && isUndefined(dataLabels.y) ? coordinate.height / 2 : 0),
                        y0 = defined(dataLabels.y) ? coordinate.XY[1] : isPyramid && isLeftHalf(idx) ? d.leftY : d.y,
                        polygon = [
                            [getStartX(d, idx, coordinate), y0],
                            [getEndX(d, idx, overlap, 0, coordinate), y0],
                            [getEndX(d, idx, overlap, undefined, coordinate), y]
                        ];

                    return d3Line(polygon);
                };

                // add to drawing queue
                var q = queue(commonProperties, index, subSeriesIndex, null, null, rendererIndex);
                objs.forEach(_ => (_.queue = q));
                labelArray.push(...objs);
                labelArray.queue = q;
                labelArray.queues = labelArray.queues || [];
                labelArray.queues.push(q);
                labelArray.subSeriesIndex = subSeriesIndex;

                q.properties = {
                    type: PIE_CHARTNAME,
                    dataLabels: dataLabels,
                    dataUpdate: dataUpdate,
                    showlabel: showlabel,
                    rp: rp,
                    trans: args[9],
                    txtHgt: txtHgt,
                    reverseColor: true,
                    onOverlap: function(_) {
                        _.overlap.stage = 'overlap';
                    },
                    contentType: contentType,
                    isMultiLine: isMultiLine,
                    chartType: cType
                };

                (labelArray.TOP = chartTop - args[9][1]), (labelArray.BOTTOM = chartHeight + args[9][1]);
                labelArray.actualType = FUNNEL_CHARTNAME;
            });

            function isdisabled(i) {
                var dt = data[data_length - 1 - i];
                var isWithinClrRange =
                    isWithinColorRange(dataset.getClr(dt), chartObj) && !isSeriesFiltered(chartObj, paneid, data_length - 1 - i);
                return dt.disabled || !dataset.getY(dt, yaxiscolumnorder) || !isWithinClrRange;
            }

            function isLeftHalf(idx) {
                switch (showlabel) {
                    case 'singleside':
                        return false;
                    case 'doubleside':
                        return idx % 2 == 0;
                    case 'strokeddoubleside':
                        return idx >= noofNotNullData / 2;
                }
            }

            function getStartX(d, i, coordinate) {
                if (defined(dataLabels.y)) {
                    let y = coordinate.XY[1];
                    return d._x(y, isLeftHalf(i));
                }

                return isLeftHalf(i) ? d.leftx + padding : d.rightx - padding;
            }

            function getEndX(d, i, options, pad, coordinate) {
                var isLeft = isLeftHalf(i);

                if (defined(dataLabels.x) && coordinate) {
                    return coordinate.XY[0];
                }

                if (!relativeX || options.stage == 'overlap') {
                    pad = defined(pad) ? pad : padding;
                    return isLeft ? 0 - pad : chartWidth + pad;
                } else {
                    return isLeft ? d.leftx - boundBoxPadding : d.rightx + boundBoxPadding;
                }
            }
        });
    }

    labels.renderer = function(_) {
        if (!arguments.length) {
            return renderer;
        }
        renderer = _;
        return labels;
    };
    labels.args = function(_) {
        if (!arguments.length) {
            return args;
        }
        args = _;
        return labels;
    };
    return labels;
}

LabelRegistry.add('funnel', funnelLabel);
export var _funnelDatalabel = funnelLabel;
