// @zohocharts/zc/src/label/funnelinnerlabels.js
//$Id$
//funnelinnerlabels.js
import { LabelRegistry } from './label-registry';
import { NULL } from '../core/zc-constants';
import { getMergedDatalabelOptions, applyDatalabelFonts } from './datalabelsUtils';
import { isBoundBoxEnabled, getBoundBoxPadding, getBoundBoxStrokeWidth } from './boundBoxUtils';
import { bbox, pick } from '../utils/zcutils';
import { dataLabelTextContent } from './datalabel-content';
import { isWithinColorRange, isSeriesFiltered } from '../utils/colorutils';
import { INNERLABEL_LABELTYPE, DATALABELS_LABELTYPE } from './labeltype-constants';
import { appendEle } from '../utils/zcsvgutils';
import { datacount_based_intelligence } from '../utils/intelligence-utils';
import { PYRAMID_CHARTNAME, PIE_CHARTNAME } from '../core/chartname-constants';
import { getFontSizeByHierarchy } from '../utils/zcdatautils';
import { translateTspan } from './tspan-utils';
import { getSvgTransform } from '../utils/svg-transforms';
import { getChartTypeByIndex } from '../core/charttype-helpers';

function funnelInnerLabel() {
    var renderer, args;

    function labels(selection, labelHandler) {
        selection.each(function() {
            var currentdata = args[0],
                index = args[1],
                data = args[2],
                arcHeight = args[4],
                totalEngagement = args[5],
                cfg = args[6],
                labelPosition = args[7];
            var chartWidth = cfg.width,
                rp = renderer.commonRendererProp().params,
                { rendererIndex = index } = rp,
                chartObj = rp.chartObj,
                commonPlotOptions = rp.commonPlotOptions,
                dataset = rp.dataset,
                yaxiscolumnorder = rp.yaxiscolumnorder,
                paneid = labelHandler.paneid,
                { name: chartName, value: cType } = getChartTypeByIndex(chartObj, paneid, index);
            var translation = `translate(${args[9]})`,
                isPyramid = chartName === PYRAMID_CHARTNAME;

            currentdata = currentdata[index];
            var plotoptions = currentdata.plotoptions;
            var dataLabels = getMergedDatalabelOptions(commonPlotOptions, plotoptions, chartObj, index);
            var innerLabels = dataLabels.innerLabel,
                padding = cfg.datalabelPadding || +dataLabels.padding || 0,
                data_length = data.length;

            var showInnerLabel = innerLabels.show === true,
                txtHgt = getFontSizeByHierarchy(innerLabels, chartObj),
                GLT = dataLabelTextContent(chartObj, rp, innerLabels, INNERLABEL_LABELTYPE, totalEngagement, index, undefined, paneid),
                contentType = GLT.contentType || 'text',
                isMultiLine = contentType === 'html' && GLT.multiLine,
                boundBoxPadding = isBoundBoxEnabled(dataLabels, chartObj)
                    ? getBoundBoxPadding(dataLabels, chartObj) + getBoundBoxStrokeWidth(dataLabels, chartObj)
                    : 0;
            showInnerLabel = showInnerLabel && datacount_based_intelligence(chartObj, DATALABELS_LABELTYPE);
            txtHgt = txtHgt + 2 * boundBoxPadding;

            if (currentdata.disabled || !showInnerLabel) {
                return;
            }

            var temp_label = appendEle(chartObj.staticPaneSVG.select('.labelholder'), 'text', [1], '.', 'class', 'temp_label');
            applyDatalabelFonts(temp_label, innerLabels, chartObj, rp.color);
            temp_label.attr('transform', translation).style('opacity', 0);
            var queue = labelHandler.add;

            currentdata.data.forEach(function(d, i) {
                var dataUpdate = labelPosition,
                    commonProperties = [],
                    subSeriesIndex = pick(args[8], i);

                dataUpdate.forEach(function(d, j) {
                    var display = !isdisabled(j),
                        X = isPyramid ? d.midX : chartWidth / 2,
                        Y = isPyramid ? d.leftY : d.y,
                        ANCHOR = 'middle',
                        DY = isMultiLine ? '0.9em' : '0.5em',
                        text = GLT(data[data_length - 1 - j], data_length - 1 - j),
                        bb = null;

                    temp_label[contentType](text).attrs({
                        x: X,
                        y: Y,
                        dy: DY,
                        'text-anchor': ANCHOR
                    });

                    if (display && isMultiLine) {
                        translateTspan(temp_label, X);
                        bb = bbox(temp_label.node());
                        Y -= bb.height / 2;
                        txtHgt = bb.height;
                    }

                    if (display) {
                        bb = bb || bbox(temp_label.node());
                        if (isPyramid) {
                            display = !(txtHgt > (d.y0 - d.leftY) * 2);
                        } else {
                            var ellipseArcHeight = cfg.canOverlap
                                ? ((d.rightx - d.leftx - 2 * padding) / chartWidth) * arcHeight
                                : arcHeight;
                            display = !(
                                txtHgt < d.y - (d.y0 - d.y) + ellipseArcHeight &&
                                txtHgt > d.y0 + ellipseArcHeight - (d.y - (d.y0 - d.y) + ellipseArcHeight)
                            );
                        }
                    }

                    var coordinate = {
                        X: X,
                        Y: Y,
                        DY: DY,
                        ANCHOR: ANCHOR,
                        text: text,
                        display: display,
                        ID: `innerlbl${index}_${subSeriesIndex}_${j}`,
                        width: display ? bb.width : 0,
                        height: display ? bb.height : 0
                    };
                    commonProperties.push(coordinate);
                });

                commonProperties.dataUpdate = dataUpdate;
                commonProperties.data = data;
                commonProperties.translation = `translate(${args[9]})`;

                // add to drawing queue
                var q = queue(commonProperties, index, subSeriesIndex, NULL, INNERLABEL_LABELTYPE, rendererIndex, paneid);
                q.properties = {
                    type: PIE_CHARTNAME,
                    dataLabels: innerLabels,
                    dataUpdate: dataUpdate,
                    showlabel: showInnerLabel,
                    rp: rp,
                    trans: args[9],
                    reverseColor: true,
                    contentType: contentType,
                    isMultiLine: isMultiLine,
                    chartType: cType,
                    data
                };
            });

            function isdisabled(i) {
                var dt = data[data_length - 1 - i];
                var isWithinClrRange =
                    isWithinColorRange(dataset.getClr(dt), chartObj) && !isSeriesFiltered(chartObj, paneid, data_length - 1 - i);
                return dt.disabled || !dataset.getY(dt, yaxiscolumnorder) || !isWithinClrRange;
            }
        });
    }

    labels.renderer = function(_) {
        if (!arguments.length) {
            return renderer;
        }
        renderer = _;
        return labels;
    };
    labels.args = function(_) {
        if (!arguments.length) {
            return args;
        }
        args = _;
        return labels;
    };
    return labels;
}

LabelRegistry.add('funnelInnerLabels', funnelInnerLabel);
export var _funnelInnerDatalabel = funnelInnerLabel;
