// @zohocharts/zc/src/label/hierarchy-datalabels.js
//$Id$
//hierarchy-datalabels.js
import { LabelRegistry } from './label-registry';
import { getPlotBound } from '../plot/plotSize';
import { getMergedDatalabelOptions, isDatalabelEnabled, applyDatalabelFonts, isLabelWithinBound, getLabelArray } from './datalabelsUtils';
import { mergeJSON, isTrue, validateNullVal, parseShortHandValue, bbox, defined, perToPx, isProperObject, invokeFunction } from '../utils/zcutils';
import {
    getMergedBoundBoxOptions,
    getBoundBoxHookConf,
    getBoundBoxPadding,
    getBoundBoxStrokeWidth,
    isBoundBoxEnabled
} from './boundBoxUtils';
import { isMultiColoring } from '../utils/colorutils';
import { getFontSizeByHierarchy, getFontFamilyByHierarchy } from '../utils/zcdatautils';
import { CENTER_ALIGN, RIGHT_ALIGN, BOTTOM_ALIGN, START_ALIGN, END_ALIGN, MIDDLE_ALIGN, LEFT_ALIGN } from '../core/align-constants';
import { mathMax, NONE, NULL } from '../core/zc-constants';
import { d3_scaleLinear } from '../core/d3-constants';
import { dataLabelTextContent } from './datalabel-content';
import { appendEle } from '../utils/zcsvgutils';
import { DATALABELS_LABELTYPE } from './labeltype-constants';
import { isWithinColorRange, isSeriesFiltered } from '../utils/colorutils';
import { colorscale_helpers } from '../axis/colorscale-helpers';
import { PACKEDBUBBLE_CHARTNAME, HEATMAP_CHARTNAME, TREEMAP_CHARTNAME, SANKEY_CHARTNAME } from '../core/chartname-constants';
import { getHeatCellPosition, getPackedbubblePosition } from '../events/events-utils';
import { getTextBound } from '../axis/textBound';
import { getTspanBound, trimAllTspan, translateTspan, getCustomHTML } from './tspan-utils';
import { callTrimText } from './datalabelsUtils';
import { mathMin, mathPow, mathSQRT } from '../core/zc-constants';
import { PARENTLABELS_LABELTYPE } from './labeltype-constants';
import { treemapUtils } from '../treemap/treemap-utils';
import { getDy4MultiLine } from './position-helpers';
import { getChartTypeByIndex } from '../core/charttype-helpers';
import { getRendererIndex } from '../core/data-parser';

/*
 * PENDING:
 * Bound box - minHeight should be considered for height on handling overlapping datalabels (if hook is enabled)
 */
//multi coloring
function hierarchyLabel() {
    var renderer, args, labelseries;

    function labels(selection, labelHandler) {
        selection.each(function() {
            var currentdata = args[0],
                index = args[1],
                size = args[2] || 0,
                yrangePos = args[3],
                total = args[4],
                leafNodes = args[5],
                transform = args[6] || [0, 0],
                isParent = args[7],
                isPreCalculating = args[8],
                plotoptions = currentdata.plotoptions,
                dataLabels,
                paneid = labelHandler.paneid,
                rp = renderer.commonRendererProp().params,
                chartObj = rp.chartObj,
                systemconf = chartObj.systemConf,
                commonPlotOptions = rp.commonPlotOptions,
                dataset = rp.dataset,
                yaxiscolumnorder = rp.yaxiscolumnorder,
                labelArray = getLabelArray(rp, paneid),
                xscale = rp.xscale,
                yscale = rp.yscale,
                { rendererIndex = index } = rp;

            dataLabels = getMergedDatalabelOptions(commonPlotOptions, plotoptions, chartObj, index);
            var showlabel = isDatalabelEnabled(dataLabels, chartObj),
                cType_obj = getChartTypeByIndex(chartObj, paneid, 0),
                chartType = cType_obj.value,
                chartName = cType_obj.name,
                isSankeyChart = chartName === SANKEY_CHARTNAME,
                sankeyLabelInside = isSankeyChart && dataLabels.labelPos === 'inside',
                maxWidth = perToPx(defined(dataLabels.maxWidth) ? dataLabels.maxWidth : '30%', chartObj.plotarea.width);

            if (isParent) {
                var parentLabelOptions = mergeJSON(commonPlotOptions.parent.datalabels, dataLabels);
                showlabel = isTrue(parentLabelOptions.show);
                dataLabels = parentLabelOptions;
            }

            if (currentdata.disabled || !showlabel || isSeriesFiltered(chartObj, paneid, index)) {
                return;
            }

            var currentPlotBound = getPlotBound(chartObj),
                boundBoxPadding = 0,
                boundBoxStrokeWidth = 0,
                hookHeight = 0,
                hookWidth = 0,
                boundBoxRadius = 0;
            if (isBoundBoxEnabled(dataLabels, chartObj)) {
                var boundBoxOptions = getMergedBoundBoxOptions(dataLabels, chartObj);
                var hookOptions = getBoundBoxHookConf(boundBoxOptions.hook);
                boundBoxPadding = getBoundBoxPadding(dataLabels, chartObj);
                boundBoxStrokeWidth = getBoundBoxStrokeWidth(dataLabels, chartObj);
                hookHeight = hookOptions.height;
                hookWidth = hookOptions.width;
                boundBoxRadius = boundBoxOptions.radius;
            }

            var textOverlap = dataLabels.textOverlap, //dataLabels.textOverlap, //hidden | adjust | ellipsis
                hAlign = dataLabels.hAlign || CENTER_ALIGN,
                vAlign = dataLabels.vAlign || CENTER_ALIGN,
                padding = validateNullVal(dataLabels.padding, 0.5),
                handleOverlapping = systemconf.chart.plot.datalabels.handleOverlapping === true && textOverlap !== NONE,
                GLT = dataLabelTextContent(chartObj, rp, dataLabels, DATALABELS_LABELTYPE, total, index, yrangePos, paneid),
                contentType = GLT.contentType || 'text',
                isMultiLine = contentType === 'html' && GLT.multiLine,
                hasShowAsFunction = GLT.hasShowAsFunction,
                multiColoring = isMultiColoring(commonPlotOptions, chartObj),
                fontSize = getFontSizeByHierarchy(dataLabels, chartObj),
                heightAdjustScale = d3_scaleLinear().domain([-0.3, 0.9]),
                isCanvasMode = chartObj.plot.renderer.mode === 'canvas',
                globalDY = getDY(),
                fontStyles = applyDatalabelFonts(null, dataLabels, chartObj, rp.color, false, index);

            textOverlap = hasShowAsFunction && textOverlap === 'ellipsis' ? 'hidden' : textOverlap;

            var temp_label = appendEle(
                chartObj.staticPaneSVG.select('.labelholder'),
                hasShowAsFunction ? 'g' : 'text',
                [1],
                '.',
                'class',
                'temp_label'
            );
            var updateHandler = labelHandler.update,
                queue = labelHandler.add;

            // currentdata.forEach(function(d, i) {
            var dataUpdate = leafNodes;

            var calculated_points = [];
            applyDatalabelFonts(temp_label, dataLabels, chartObj, rp.color, false, index);
            temp_label.style('opacity', 0);

            dataUpdate.forEach(function(d, j) {
                var ANCHOR = getAnchor(d),
                    pos = getPosition(d, ANCHOR);
                var x = pos.x,
                    y = pos.y,
                    text = null,
                    tspanInfo = null;
                var display = colorscale_helpers.isQuantileColorScale(chartObj)
                        ? !isParent
                            ? isWithinColorRange(dataset.getClr(d.data || d), chartObj)
                            : true
                        : true,
                    id = '',
                    dy,
                    actualDY,
                    HOOKPOSITION = null,
                    dy,
                    actualDY,
                    xDiff = 0,
                    yDiff = 0,
                    polarXY;

                if (isSankeyChart) {
                    if (ANCHOR === START_ALIGN) {
                        HOOKPOSITION = LEFT_ALIGN;
                    } else if (ANCHOR === END_ALIGN) {
                        HOOKPOSITION = RIGHT_ALIGN;
                    }
                }

                temp_label[contentType](function() {
                    text = GLT(d.data || d, j, pos);
                    return text;
                });

                if (isMultiLine) {
                    let t = getDy4MultiLine(temp_label, dataLabels, globalDY, text, display);
                    dy = `${t[0]}em`;
                    actualDY = t[1];
                    text = t[2];
                    display = t[3];
                } else {
                    dy = `${globalDY}em`;
                }

                temp_label.attrs({
                    dy: dy,
                    transform: `translate(${[x, y]})`,
                    'text-anchor': ANCHOR,
                    display: display
                });

                var textBound;
                if (contentType === 'html') {
                    //Test_665.json
                    if (isMultiLine) {
                        translateTspan(temp_label, 0);
                    }

                    if (hAlign !== CENTER_ALIGN && !hasShowAsFunction) {
                        //ZC1494: If font-weight is bold or font-style is italic, left/right is out of bound
                        temp_label.attr('text-anchor', END_ALIGN);
                        textBound = bbox(temp_label);
                        var dEnd = textBound.width + textBound.x;

                        temp_label.attr('text-anchor', START_ALIGN);
                        textBound = bbox(temp_label);
                        var dStart = textBound.x;

                        x += (dStart - dEnd) * (hAlign === RIGHT_ALIGN ? 1 : -1);

                        temp_label.attr('transform', `translate(${[x, y]})`).attr('text-anchor', ANCHOR);
                    }

                    textBound = bbox(temp_label);
                } else {
                    textBound = getTextBound(
                        text,
                        fontSize,
                        getFontFamilyByHierarchy(dataLabels, chartObj),
                        null,
                        null,
                        fontStyles['font-style'],
                        fontStyles['font-weight']
                    );
                    textBound = updateTextBound(textBound, ANCHOR, dy, x, y);
                }

                var widthToBeAdded = 0;
                var bb_width = textBound.width + widthToBeAdded,
                    bb_height = textBound.height,
                    bb_x = textBound.x + widthToBeAdded / 2;

                if (handleOverlapping) {
                    // update the width with new label width
                    // Consider padding for label bound
                    var labelBound = {
                            x1: textBound.x + x - boundBoxPadding,
                            y1: textBound.y + y - boundBoxPadding,
                            x2: textBound.x + x + textBound.width + boundBoxPadding,
                            y2: textBound.y + y + textBound.height + boundBoxPadding
                        },
                        rawPosition = getRawPosition(d, textBound);

                    let nodeBound = getNodeBound(rawPosition, textBound, labelBound);

                    if (isSankeyChart) {
                        pos.w = nodeBound.x2 - nodeBound.x1;
                        pos.h = nodeBound.y2 - nodeBound.y1;
                    }

                    var within = isLabelWithinBound(labelBound, nodeBound);

                    if (!within) {
                        var trimmed;
                        if (textOverlap === 'ellipsis') {
                            // add ellipsis
                            var h = pos.h;
                            var w = pos.w;
                            if (bb_height + 2 * boundBoxPadding < h) {
                                var bound = {
                                    left: nodeBound.x1 + boundBoxPadding,
                                    top: nodeBound.y1 + boundBoxPadding,
                                    right: nodeBound.x2 - boundBoxPadding,
                                    bottom: nodeBound.y2 - boundBoxPadding
                                };

                                if (contentType === 'html') {
                                    tspanInfo = getTspanBound(temp_label, fontSize);
                                    trimAllTspan(temp_label, chartObj, transform, 0, bound, tspanInfo, NULL, [x, y]);
                                } else {
                                    callTrimText(temp_label, chartObj, transform, 0, bound, [x, y]);
                                }

                                var trimmedText = temp_label.text();
                                if (trimmedText && trimmedText !== '..') {
                                    //If the textContent is just "..", hide the label
                                    text = temp_label[contentType]();
                                    trimmed = true;
                                    bb_width = w - 2 * boundBoxPadding;
                                    if (isMultiLine) {
                                        var newBB = bbox(temp_label);
                                        bb_height = newBB.height;
                                        bb_width = newBB.width + widthToBeAdded;
                                        y += (textBound.height - newBB.height) / 2;
                                    }
                                }
                            }
                        }
                        // if (!isParent) {
                            if (!trimmed) {
                                display = false;
                            } else if (contentType === 'html') {
                                tspanInfo = getTspanBound(temp_label, fontSize);
                            }
                        // }
                    }

                    // for sankey chart we need overlap handler
                    // ZC2593
                    if (chartName === SANKEY_CHARTNAME && !sankeyLabelInside) {
                        labelBound.x1 = mathMax(nodeBound.x1, labelBound.x1);
                        labelBound.x2 = mathMin(nodeBound.x2, labelBound.x2);

                        let obj = {
                                id: `lbl${index}_${0}_${j}`,
                                x1: labelBound.x1,
                                y1: labelBound.y1,
                                x2: labelBound.x2,
                                y2: labelBound.y2,
                                translateFunction: updateHandler,
                                overlap: 'hidden',
                                hookHeight: hookHeight,
                                hookPosition: HOOKPOSITION
                            },
                            // check if the point is out if plot
                            { left, top, right, bottom } = getPlotBound(chartObj),
                            pointOutOfPlot = labelBound.x1 < left || labelBound.x2 > right || labelBound.y1 < top || labelBound.y2 > bottom;

                        if (display && !isPreCalculating) {
                            display = !pointOutOfPlot;
                        }
                        labelArray.push(obj);
                        // Label Testing Area
                    }
                }
                let translateX, translateY;
                if (isSankeyChart && chartObj.dataObject.sankeyLayout.zoomTransform) {
                    let zoomTransform = chartObj.dataObject.sankeyLayout.zoomTransform;

                    translateX = x;
                    translateY = y;

                    x = (x * parseFloat(zoomTransform.k)) + parseFloat(zoomTransform.x);
                    y = (y * parseFloat(zoomTransform.k)) + parseFloat(zoomTransform.y);
                }

                var coordinate = {
                    X: 0,
                    Y: 0,
                    DY: dy,
                    xy: [x, y],
                    _translate: [translateX, translateY],
                    actualDY: actualDY,
                    ANCHOR: ANCHOR,
                    text: text,
                    display: display,
                    height: bb_height,
                    width: bb_width,
                    hookPosition: HOOKPOSITION,
                    adjustVal: xDiff
                };

                if (isCanvasMode && contentType === 'html') {
                    if (hasShowAsFunction) {
                        coordinate.customHTML = getCustomHTML(temp_label);
                    } else {
                        coordinate.tspanInfo = tspanInfo || getTspanBound(temp_label, fontSize);
                    }
                }

                calculated_points.push(coordinate);
            });

            calculated_points.dataUpdate = dataUpdate; //Each subseries may have different dataUpdates
            var q = queue(calculated_points, index, 0, yrangePos, isParent ? PARENTLABELS_LABELTYPE : NULL, rendererIndex);
            q.properties = {
                type: 'hierarchy',
                dataLabels: dataLabels,
                color: rp.color,
                dataUpdate: dataUpdate,
                showlabel: showlabel,
                multiColoring: multiColoring,
                rp: rp,
                contentType: contentType,
                isMultiLine: isMultiLine,
                GLT: GLT,
                trans: [0, 0],
                chartType
            };
            // });

            function getPosition(d, ANCHOR) {
                var pos = getRawPosition(d),
                    x = pos.x,
                    y = pos.y,
                    w = pos.width,
                    h = pos.height;

                x = updateXposition(x, w, d, ANCHOR) + transform[0];
                y = updateYposition(y, h) + transform[1];

                return {
                    x: x,
                    y: y,
                    w: w,
                    h: h
                };
            }

            function getNodeBound(pos, textBound, labelBound) {
                var nodeheight = pos.height - 2 * padding,
                    nodewidth = pos.width - 2 * padding,
                    nodex = pos.x + transform[0] + padding,
                    nodey = pos.y + transform[1] + padding;

                var bb = {
                    x1: nodex,
                    y1: nodey,
                    x2: nodex + nodewidth,
                    y2: nodey + nodeheight
                };

                // when the sankey node is smaller than text height, center by text bound
                if (chartName === SANKEY_CHARTNAME) {
                    bb.y1 = labelBound.y1 - 2;
                    bb.y2 = labelBound.y2 + 2;

                    //bbox floating issue in Windows, Ubuntu, Mac - bb.x is not equal to bb.width/2 - CRM team Zoho - CommandCenter
                    bb.x2 += 2;
                }

                if (chartName === PACKEDBUBBLE_CHARTNAME) {
                    // var radius = pos.w / 2,
                    //     chord = mathMin(textBound.height / 2 + boundBoxPadding, radius),
                    //     apothem = mathSQRT(mathPow(radius, 2) - mathPow(chord * 2, 2) / 4);
                    // excess = radius - apothem;
                    var radius = nodewidth / 2, // hypotenuse
                        vert = mathMin(textBound.height / 2, radius),
                        hori = mathSQRT(mathPow(radius, 2) - mathPow(vert, 2)),
                        excess = radius - hori;
                    bb.x1 += excess;
                    bb.x2 -= excess;
                    bb.y1 += excess;
                    bb.y2 -= excess;
                }
                return bb;
            }

            function updateXposition(x, w, d, ANCHOR) {
                if (hAlign === CENTER_ALIGN) {
                    x += w / 2;
                } else if (hAlign === RIGHT_ALIGN) {
                    x += w - padding - boundBoxPadding;
                } else {
                    x += padding + boundBoxPadding;
                }
                return x;
            }

            function updateYposition(y, h) {
                if (vAlign === CENTER_ALIGN) {
                    y += h / 2;
                } else if (vAlign === BOTTOM_ALIGN) {
                    y += h - padding - boundBoxPadding;
                } else {
                    y += padding + boundBoxPadding;
                }
                return y;
            }

            function updateSankeyXPosition(x, w, d, ANCHOR) {
                if (ANCHOR === MIDDLE_ALIGN || sankeyLabelInside) {
                    x += w / 2;
                } else if (ANCHOR === END_ALIGN) {
                    x -= padding + boundBoxPadding;
                } else {
                    x += padding + boundBoxPadding;
                }
                return x;
            }

            function getAnchor(d) {
                if (isSankeyChart) {
                    return renderer.getLabelAnchor(d);
                }
                return hAlign === CENTER_ALIGN ? 'middle' : hAlign === RIGHT_ALIGN ? 'end' : null;
            }

            function getDY() {
                var DY;
                if (vAlign === CENTER_ALIGN) {
                    DY = 0.3;
                } else if (vAlign === BOTTOM_ALIGN) {
                    DY = -0.3;
                } else {
                    DY = 1;
                }

                return DY;
            }

            function getRawPosition(d, textBound) {
                switch (chartName) {
                    case PACKEDBUBBLE_CHARTNAME:
                        return getPackedbubblePosition(d);
                    case TREEMAP_CHARTNAME:
                        var layout = chartObj.renderer[getRendererIndex(d.data, chartObj)].layout;
                        var pos = treemapUtils.getTileBound(layout, d);
                        pos = adjustTileBoundForParents(pos, d);
                        return pos;
                    case HEATMAP_CHARTNAME:
                        return getHeatCellPosition(chartObj, d, xscale, yscale, yaxiscolumnorder);
                    case SANKEY_CHARTNAME: {
                        return renderer.getPosition(d, textBound, padding, maxWidth, boundBoxPadding, hookHeight);
                    }
                }
            }

            function updateTextBound(textBound, ANCHOR, dy, x, y) {
                var w = textBound.width,
                    h = textBound.height;
                var tx = ANCHOR === 'middle' ? w / 2 : ANCHOR === 'end' ? w : 0,
                    ty = heightAdjustScale.range([h, 0])(parseFloat(dy));
                textBound.x = -tx;
                textBound.y = -ty;
                return textBound;
            }

            function adjustTileBoundForParents(pos, d) {
                if (isParent && (dataLabels.position === 'bottom-outset' || dataLabels.position === 'top-outset')) {
                    var padding = commonPlotOptions.padding;
                    var tilePadding = isProperObject(padding) ? padding : parseShortHandValue(padding),
                        adjust = tilePadding[dataLabels.position === 'bottom-outset' ? 'bottom' : 'top'];

                    adjust = invokeFunction(adjust, d);  // #ZC4723 - callback support for treemap.padding
                    if (dataLabels.position === 'bottom-outset') {
                        pos.y = pos.y + pos.height - adjust;
                    }
                    pos.height = adjust <= pos.height ? adjust : pos.height;
                    //inorder to avoid heightwise adjustment properly we do above.
                    //previously adjust was sent directly, if height is less than adjust - the bound taken was higher than actual, so if only if adjust is less than height we assign it.
                }
                return pos;
            }
        });
    }

    labels.renderer = function(_) {
        if (!arguments.length) {
            return renderer;
        }
        renderer = _;
        return labels;
    };
    labels.args = function(_) {
        if (!arguments.length) {
            return args;
        }
        args = _;
        return labels;
    };

    return labels;
}

LabelRegistry.add('hierarchy', hierarchyLabel);
export var _hierarchyDatalabel = hierarchyLabel;
