// @zohocharts/zc/src/label/labelrenderer.js
//$Id$
//labelrenderer.js
import { Registry } from '../core/zc-registry';
import { LabelRegistry } from './label-registry';
import { STACKLABELS_LABELTYPE, PARENTLABELS_LABELTYPE, INNERLABEL_LABELTYPE, DATALABELS_LABELTYPE } from './labeltype-constants';
import { AREA_CHARTNAME, BAR_CHARTNAME } from '../core/chartname-constants';
import { dataLabel_handler } from '../core/main-zc-init';
import { isSunburstChartType } from '../events/events-utils';
import { noOperation } from '../core/zc-constants';
import { each, arrayIncludes, defined, getDistinct } from '../utils/zcutils';
import { createElement } from '../utils/zcsvgutils';
import { isFacet } from '../axis/hierarchy-axis-utils';
import { getSeriesIndex4StackLabel } from '../core/data-processor/transforms/stack/stack-grouping';
import { getChartTypeByIndex, isCategoryOf } from '../core/charttype-helpers';
import { pointTravel } from './overlapHandler/point-travel';
import { stepTravel } from './overlapHandler/step-travel';
import { slotTravel } from './overlapHandler/slot-travel';
import { DataProcessor } from '../core/data-processor';
import { getAllSeriesTypes } from '../core/series-parser-helpers';

export * from './datalabels-precalculation';
export * from '../scale/evaluators/correction/label-evaluator';
export * from './datalabels-init';

function LabelRenderer(labelQueue, array, chartObj) {
    chartObj.labelHandler = null;
    if (chartObj.plot.renderer.mode !== 'image') {
        var dataObj = chartObj.dataObject;

        if (chartObj.isAxisCategory) {
            drawStackLabel(chartObj);
        }

        // manage rendering order of datalabels 
        var isComboChart = dataObj.isCombinationalChart;
        var allChartTypes = isComboChart? chartObj.distinctChartTypes || getDistinct(getAllSeriesTypes(chartObj.userdata)) : [];
        var renderingOrder = allChartTypes.includes(AREA_CHARTNAME) && allChartTypes.includes(BAR_CHARTNAME) ? [BAR_CHARTNAME] : [],
            renderingOrderLength = renderingOrder.length;
        if (renderingOrderLength) {
            var renderingOrderAsObject = {},
                globalChartType = chartObj.dataObject.globalChartType,
                getChartTypeString = function (chartType) { return DataProcessor.helpers.getChartType(chartType, "string") };

            renderingOrder.forEach(function (cType, i) { return renderingOrderAsObject[getChartTypeString(cType)] = i; });

            labelQueue.sort(function (type1, type2) {
                const type1Order = renderingOrderAsObject[getChartTypeString(type1.args[0].type || globalChartType)],
                    type2Order = renderingOrderAsObject[getChartTypeString(type2.args[0].type || globalChartType)],
                    comparator = (defined(type1Order) ? type1Order : renderingOrderLength) - (defined(type2Order) ? type2Order : renderingOrderLength);
                return comparator;
            })
        }

        labelQueue.forEach(drawLabel);

        //handle overlap - start
        for (var attr in array) {
            getOverlapHandlerType(chartObj, attr)(chartObj, array[attr], labelQueue);
        }

        //handle overlap - end

        // remove the temp node.
        chartObj.staticPaneSVG.select('.temp_label').remove();

        if (chartObj.labelHandler) {
            var labelHolder = chartObj.staticPaneSVG.selectAll('g.labelholder'),
                mode = chartObj.plot.renderer.mode,
                hierarchical = chartObj.dataObject.hierarchical,
                facetRendering = hierarchical && hierarchical.facetRendering;

            labelQueue.forEach(function(d) {
                var legendOrder = d.args[1],
                    rendererIndex = d.rendererIndex,
                    paneid = d.paneid,
                    innerLabels = /(pieInnerLabels|funnelInnerLabels)/.test(d.charttype),
                    labelType =
                        d.charttype === 'pieInnerLabels' &&
                        isSunburstChartType(getChartTypeByIndex(chartObj, paneid, legendOrder).value) &&
                        d.args[d.args.length - 1]
                            ? PARENTLABELS_LABELTYPE
                            : innerLabels
                            ? INNERLABEL_LABELTYPE
                            : DATALABELS_LABELTYPE;

                
                var queue = findQueue(d, chartObj);

                let boundBoxComponent = LabelRegistry.get(`boundBox-${mode}`);
                if (boundBoxComponent !== noOperation) {
                    var datalabelsBoundBox_renderer = boundBoxComponent().args(queue),
                        boundBoxWrap =
                            facetRendering && defined(paneid) ? createElement('g', `.grid${paneid}`, [1], labelHolder) : labelHolder;

                    boundBoxWrap.call(datalabelsBoundBox_renderer, legendOrder, labelType, rendererIndex, d);
                }
            });
            chartObj.labelHandler.draw();
        }
    }
}

function drawLabel(obj) {
    var chartObj = obj.chartObj,
        renderer = obj.renderer,
        cType = obj.charttype,
        rendererIndex = obj.rendererIndex,
        paneid = obj.paneid,
        args = obj.args;
    var labelHolder = chartObj.staticPaneSVG.selectAll('g.labelholder');
    var handler = chartObj.labelHandler;

    if (!chartObj.labelHandler) {
        handler = chartObj.labelHandler = dataLabel_handler.initialize(chartObj);
        handler.chartType = cType;
    }
    handler.rendererIndex = rendererIndex;
    handler.paneid = paneid;

    let labelrenderer = LabelRegistry.get(cType);

    if (labelrenderer !== noOperation) {
        labelrenderer = labelrenderer();
        labelrenderer.renderer(renderer).args(args);
        labelHolder.call(labelrenderer, handler);
    }
}

function drawStackLabel(chartObj) {
    let dataObj = chartObj.dataObject;
    var stackEncode = dataObj.getEncode('stack') || [];

    each(stackEncode, (stackData, stackIndex) => {
        stackIndex = dataObj.pivot ? '0_0' : +stackIndex;
        var seriesIndex = getSeriesIndex4StackLabel(stackData);
        if ((!dataObj.noofBarEnabledSeries && !dataObj.isCombinationalChart) || seriesIndex == null) {
            return;
        }

        let stackLabelsData = stackEncode[stackIndex].layout.getAllExtents();

        // #ZC2063
        if (chartObj.renderVisiblePointsOnly) {
            let visibleDomain = chartObj.axes.x.scale.domain();
            stackLabelsData = stackLabelsData.filter(d0 => arrayIncludes(visibleDomain, d0[0]));
        }

        var obj = {
            renderer: chartObj.renderer[seriesIndex],
            chartObj: chartObj,
            args: [
                {
                    data: [stackLabelsData]
                },
                seriesIndex,
                STACKLABELS_LABELTYPE,
                stackIndex
            ],
            charttype: BAR_CHARTNAME,
            paneid: '0_0' //hardcoded need to change
        };
        drawLabel(obj);
    });
}

function getOverlapHandlerType(chartObj, paneid) {
    paneid = isFacet(chartObj) && paneid === '0' ? '0_0' : paneid;
    paneid = paneid.replace(/\((\d+)\)/, '');

    let dataObj = chartObj.dataObject,
        chartType = getChartTypeByIndex(chartObj, paneid, 0).value,
        isAxisCategory = isCategoryOf(chartObj, chartType, 'axis'),
        isPolarAxisCategory = isCategoryOf(chartObj, chartType, 'polarAxis');

    return (isAxisCategory && dataObj.isCombinationalChart) ||
        (isAxisCategory && !dataObj.noofBarEnabledSeries) ||
        isPolarAxisCategory ||
        chartObj.isGeoMap ||
        dataObj.isFlowChartCategory
        ? pointTravel
        : isAxisCategory && dataObj.noofBarEnabledSeries
        ? stepTravel
        : slotTravel;
}

export function findQueue(d, chartObj) {
    var legendOrder = d.args[1],
        rendererIndex = d.rendererIndex,
        paneid = d.paneid,
        chartType = getChartTypeByIndex(chartObj, paneid, 0).value,
        isAxisCategory = isCategoryOf(chartObj, chartType, 'axis'),
        yRangePos = isAxisCategory ? d.args[3] : null,
        innerLabels = /(pieInnerLabels|funnelInnerLabels)/.test(d.charttype),
        labelType =
            d.charttype === 'pieInnerLabels' &&
            isSunburstChartType(getChartTypeByIndex(chartObj, paneid, legendOrder).value) &&
            d.args[d.args.length - 1]
                ? PARENTLABELS_LABELTYPE
                : innerLabels
                ? INNERLABEL_LABELTYPE
                : DATALABELS_LABELTYPE;

    return chartObj.labelHandler.getQueue(legendOrder, yRangePos, labelType, rendererIndex, paneid);
}

Registry.setComponent('labelRenderer', LabelRenderer);
