// @zohocharts/zc/src/label/piedatalabels.js
//$Id$
//piedatalabels.js
import { LabelRegistry } from './label-registry';
import { isSunburstChartType } from '../events/events-utils';
import { isDatalabelEnabled, getMergedDatalabelOptions, applyDatalabelFonts, getLabelArray } from './datalabelsUtils';
import { dataLabelTextContent } from './datalabel-content';
import { DATALABELS_LABELTYPE } from './labeltype-constants';
import {
    isBoundBoxEnabled,
    getMergedBoundBoxOptions,
    getBoundBoxHookConf,
    getBoundBoxPadding,
    getBoundBoxStrokeWidth
} from './boundBoxUtils';
import { getFontSizeByHierarchy } from '../utils/zcdatautils';
import { utils, bbox, validateNullVal, pick, isObjectPropDefined } from '../utils/zcutils';
import { mathMax, mathPI } from '../core/zc-constants';
import { d3_scaleLinear } from '../core/d3-constants';
import { translateTspan } from './tspan-utils';
import { getSizeToBeAdded } from './boundBoxUtils';
import { PIE_CHARTNAME } from '../core/chartname-constants';
import { isWithinColorRange, hasColorColumnInNonAxis, isSeriesFiltered } from '../utils/colorutils';
import { LEFT_ALIGN, RIGHT_ALIGN } from '../core/align-constants';
import { appendEle } from '../utils/zcsvgutils';
import { getDY } from '../polaraxis/polarutils';
import { getChartTypeByIndex } from '../core/charttype-helpers';

function pieLabel() {
    var renderer, args, outline, isSunburst, chartObj, dataset, paneid;

    function labels(selection, labelHandler) {
        selection.each(function() {
            var currentdata = args[0],
                pieData = args[2],
                outerRadius = args[3],
                totalEngagement = args[4],
                outlineWithPad = args[5],
                translation = `translate(${args[8][0]},${args[8][1]})`,
                extraOptions = args[9] || {},
                rp = renderer.commonRendererProp().params;

            chartObj = rp.chartObj;
            dataset = chartObj.dataset;
            paneid = labelHandler.paneid;
            var systemconf = chartObj.systemConf,
                commonPlotOptions = rp.commonPlotOptions,
                labelArray = getLabelArray(rp, paneid, args[10]),
                index = rp.datasetIndex || args[1] || 0, //For pie & dial chart, rp.datasetIndex is undefined
                rendererIndex = pick(labelHandler.rendererIndex, index),
                dataLabels,
                chartheight = ((renderer.plotarea && renderer.plotarea()) || chartObj.plotarea).height,
                chartTop = (renderer.plotarea && renderer.plotarea() && renderer.plotarea().top) || 0,
                cType = getChartTypeByIndex(chartObj, paneid, index).value;
            isSunburst = isSunburstChartType(cType);
            currentdata = currentdata[index];
            outline = args[6];

            var handleOverlapping = systemconf.chart.plot.datalabels.handleOverlapping === true,
                datalabelPadding,
                padding,
                txtHgt,
                showlabel;

            var plotoptions = currentdata.plotoptions;
            var dataLabels = getMergedDatalabelOptions(commonPlotOptions, plotoptions, chartObj, index);
            showlabel = isDatalabelEnabled(dataLabels, chartObj);

            if (currentdata.disabled || !showlabel || (isSunburst && isSeriesFiltered(chartObj, paneid, index))) {
                return;
            }

            var type = dataLabels.type,
                lineConf = dataLabels.line,
                GLT = dataLabelTextContent(chartObj, rp, dataLabels, DATALABELS_LABELTYPE, totalEngagement, index, undefined, paneid),
                contentType = GLT.contentType || 'text',
                isHTMLContent = contentType === 'html',
                isMultiLine = isHTMLContent && GLT.multiLine,
                useBBox = isObjectPropDefined(dataLabels, 'preCalculation.useBBox', true),
                boundBoxPaddingAlone = 0,
                boundBoxPadding = 0,
                hookWidth = 0,
                boundBoxRadius = 0;

            if (isBoundBoxEnabled(dataLabels, chartObj)) {
                var boundBoxOptions = getMergedBoundBoxOptions(dataLabels, chartObj);
                var hookOptions = getBoundBoxHookConf(boundBoxOptions.hook);
                boundBoxPaddingAlone = getBoundBoxPadding(dataLabels, chartObj);
                boundBoxPadding = boundBoxPaddingAlone + getBoundBoxStrokeWidth(dataLabels, chartObj) + hookOptions.height;
                hookWidth = hookOptions.width;
                boundBoxRadius = boundBoxOptions.radius;
            }

            showlabel = showlabel ? type : 'hidden';
            datalabelPadding = (extraOptions.padding || +(dataLabels.padding || 0)) + boundBoxPadding;
            txtHgt = getFontSizeByHierarchy(dataLabels, chartObj);
            padding = +(utils.getValueOfPlotOptions(plotoptions, commonPlotOptions)('outerPadding') || 0);
            (renderer.autoLabelDone = false),
                (renderer.showlabel = showlabel),
                (renderer.lineStrokeWidth = validateNullVal(lineConf.strokeWidth, 1));

            var temp_label = appendEle(chartObj.staticPaneSVG.select('.labelholder'), 'text', [1], '.', 'class', 'temp_label');
            temp_label.attr('transform', translation);
            var queue = labelHandler.add;

            (isSunburst ? currentdata : hasColorColumnInNonAxis(chartObj) ? [currentdata] : currentdata.data).forEach(function(d, i) {
                var dataUpdate = renderer.seriesdataAfterUpdate ? renderer.seriesdataAfterUpdate[i] : isSunburst ? d : pieData,
                    subSeriesIndex = pick(args[10], i),
                    objs = [];

                var commonProperties = [];
                applyDatalabelFonts(temp_label, dataLabels, chartObj, rp.color);
                temp_label.style('opacity', 0);

                dataUpdate.forEach(function(d, j) {
                    var text,
                        bb = null,
                        display = !isdisabled(d),
                        heightToBeAdded = 0;

                    var angle = midAngle(d);
                    var isLeft = angle < 0 || angle > mathPI;
                    var X =
                            showlabel == 'doubleside'
                                ? angle < 0 || angle > mathPI
                                    ? -(outerRadius + datalabelPadding)
                                    : outerRadius + datalabelPadding
                                : outlineWithPad.centroid(d)[0],
                        Y =
                            outlineWithPad.centroid(d)[1] +
                            mathMax(0, txtHgt - padding - 10 - (outerRadius + outlineWithPad.centroid(d)[1])),
                        ANCHOR = isLeft ? 'end' : 'start',
                        DY = isMultiLine
                            ? '0.9em'
                            : showlabel == 'circular' && angle > mathPI / 2 && angle < (3 * mathPI) / 2
                            ? `${getDY(angle)}em`
                            : '0.3em';
                    var actualDY = DY;

                    var dt = d.data;
                    var toNull = isSunburst ? d.depth == 0 : isdisabled(d);
                    text = toNull ? null : GLT(dt, j);

                    temp_label[contentType](text).attrs({
                        x: X,
                        y: Y,
                        dy: DY,
                        'text-anchor': ANCHOR
                    });

                    if (display && isMultiLine) {
                        translateTspan(temp_label, X);
                        var bb = bbox(temp_label.node());
                        if (showlabel == 'circular') {
                            var yscale = d3_scaleLinear()
                                .range([-1, 0])
                                .domain(angle < mathPI ? [0, mathPI] : [2 * mathPI, mathPI]);
                            Y += yscale(angle) * bb.height;
                            X += (angle > 0 && angle < mathPI ? 0.5 : angle > mathPI && angle < 2 * mathPI ? -0.5 : 0) * bb.width;
                        } else {
                            Y -= 0.5 * bb.height;
                            if (!useBBox) {
                                X -= (bb.width / 2) * (isLeft ? 1 : -1);
                            }
                        }

                        temp_label.attr('y', Y);
                        translateTspan(temp_label, X);
                    }

                    if (handleOverlapping) {
                        renderer.showlabel = 'auto';
                        var MARGIN = 5,
                            extra = 10;

                        if (!isMultiLine) {
                            DY = '0.3em';
                            temp_label.attr('dy', DY); // Test_495.json
                        }

                        if (display) {
                            bb = bbox(temp_label.node());
                            heightToBeAdded = getSizeToBeAdded(bb.height, boundBoxPaddingAlone, boundBoxRadius, hookWidth);
                            var DX = isLeft ? -(outerRadius + datalabelPadding + extra) : outerRadius + datalabelPadding + extra;
                            if (isMultiLine && !useBBox) {
                                DX += (bb.width / 2) * (isLeft ? -1 : 1);
                            }

                            objs.push({
                                id: `lbl${index}_${i}_${j}`,
                                x1: bb.x - boundBoxPadding,
                                x2: bb.x + bb.width + boundBoxPadding,
                                y1: bb.y - MARGIN - boundBoxPadding - heightToBeAdded / 2,
                                y2: bb.y + bb.height + MARGIN + boundBoxPadding + heightToBeAdded / 2,
                                isLeft: isLeft,
                                actualDY: actualDY,
                                DX: DX - X,
                                overlappingHookPosition: isLeft ? RIGHT_ALIGN : LEFT_ALIGN
                            });
                        }
                    } else {
                        bb = bbox(temp_label.node());
                        heightToBeAdded = getSizeToBeAdded(bb.height, boundBoxPaddingAlone, boundBoxRadius, hookWidth);
                    }

                    var coordinate = {
                        X: X,
                        Y: Y,
                        DY: DY,
                        ANCHOR: ANCHOR,
                        text: text,
                        display: display,
                        ID: `lbl${index}_${subSeriesIndex}_${j}`,
                        width: display ? bb.width : 0,
                        height: display ? bb.height + heightToBeAdded : 0,
                        hookPosition: showlabel === 'circular' ? null : isLeft ? RIGHT_ALIGN : LEFT_ALIGN
                    };

                    if (isHTMLContent) {
                        coordinate.isLeft = isLeft;
                    }

                    commonProperties.push(coordinate);
                });

                // add to drawing queue
                commonProperties.id = `${paneid}(${index}_${subSeriesIndex})`;
                commonProperties.translation = translation;
                commonProperties.dataUpdate = dataUpdate;
                commonProperties.renderer = renderer;

                var q = queue(commonProperties, index, subSeriesIndex, null, null, rendererIndex);
                objs.forEach(_ => (_.queue = q));
                labelArray.push(...objs);
                labelArray.queue = q;
                labelArray.queues = labelArray.queues || [];
                labelArray.queues.push(q);
                labelArray.subSeriesIndex = subSeriesIndex;

                q.properties = {
                    type: PIE_CHARTNAME,
                    dataLabels: dataLabels,
                    dataUpdate: dataUpdate,
                    showlabel: renderer.showlabel,
                    rp: rp,
                    trans: args[8],
                    txtHgt: txtHgt,
                    isSunburst: isSunburst,
                    connectorPath: function(d, i) {
                        var angle = midAngle(d);
                        var str = `M ${outline.centroid(d)[0]} ${outline.centroid(d)[1]} L${outlineWithPad.centroid(d)[0]} ${
                            outlineWithPad.centroid(d)[1]
                        }`;
                        return `${str} L${outlineWithPad.centroid(d)[0]} ${outlineWithPad.centroid(d)[1]} L${
                            angle < 0 || angle > mathPI
                                ? -(outerRadius + datalabelPadding - boundBoxPadding)
                                : outerRadius + datalabelPadding - boundBoxPadding
                        } ${outlineWithPad.centroid(d)[1]}`;
                    },
                    onNonOverlap: function(_queue) {
                        _queue.showlabel = _queue.renderer.showlabel = showlabel;
                        _queue.renderer.autoLabelDone = true;
                    },
                    onOverlap: function(_queue) {
                        _queue.connectorPath = function(d, i) {
                            // console.log(_queue, q[0]);
                            var extra = 10;
                            var angle = midAngle(d);
                            var pad = mathMax(0, txtHgt - padding - 10 - (outerRadius + outlineWithPad.centroid(d)[1]));
                            var y = outlineWithPad.centroid(d)[1] + pad;
                            var flip = angle < 0 || angle > mathPI ? -1 : 1;
                            var newY = _queue[i].Y + (isMultiLine ? _queue[i].height / 2 : 0);
                            // var path = "M " + outline.centroid(d)[0] + " " + outline.centroid(d)[1] +
                            //     " L" + outlineWithPad.centroid(d)[0] + " " + outlineWithPad.centroid(d)[1] +
                            //     " L" + ( (outerRadius + datalabelPadding) * flip ) + " " + outlineWithPad.centroid(d)[1] +
                            //     " L" + ( (outerRadius + datalabelPadding + extra) * flip ) + " " + newY;
                            // ele
                            // .attr("d", path);

                            var xControl =
                                    newY - y
                                        ? (outerRadius + datalabelPadding - boundBoxPadding + extra) * flip
                                        : outlineWithPad.centroid(d)[0],
                                yControl = outlineWithPad.centroid(d)[1],
                                x2 = (outerRadius + datalabelPadding - boundBoxPadding + extra) * flip,
                                y2 = newY;
                            var path = `M ${outline.centroid(d)[0]} ${outline.centroid(d)[1]} Q ${xControl} ${yControl} ${x2} ${y2}`;
                            return path;
                        };
                    },
                    contentType: contentType,
                    isMultiLine: isMultiLine,
                    chartType: cType
                };

                (labelArray.TOP = chartTop - args[8][1]), (labelArray.BOTTOM = chartTop + chartheight - args[8][1]);
            });
        });
    }

    labels.renderer = function(_) {
        if (!arguments.length) {
            return renderer;
        }
        renderer = _;
        return labels;
    };
    labels.args = function(_) {
        if (!arguments.length) {
            return args;
        }
        args = _;
        return labels;
    };

    function isdisabled(d) {
        var isWithinClrRange = isWithinColorRange(dataset.getClr(isSunburst ? d : d.data), chartObj);
        let active = isWithinClrRange && (isSunburst ? true : !isSeriesFiltered(chartObj, paneid, d.index));
        return !active || (isSunburst ? d.parent == null || d.value == 0 : d.value == 0);
    }

    function midAngle(d) {
        var start = isSunburst ? outline.startAngle()(d) : d.startAngle,
            end = isSunburst ? outline.endAngle()(d) : d.endAngle;
        return (start + end) / 2;
    }

    return labels;
}

LabelRegistry.add('pie', pieLabel);
export var _pieDatalabel = pieLabel;
