// @zohocharts/zc/src/label/pieinnerlabels.js
//$Id$
//pieinnerlabels.js
import { LabelRegistry } from './label-registry';
import { chartcategory } from '../core/main-zc-init';
import { NULL, NONE } from '../core/zc-constants';
import { getMergedDatalabelOptions, applyDatalabelFonts } from './datalabelsUtils';
import { datacount_based_intelligence } from '../utils/intelligence-utils';
import { isSunburstChartType } from '../events/events-utils';
import { defined, mergeJSON, bbox, pick } from '../utils/zcutils';
import { dataLabelTextContent } from './datalabel-content';
import { isBoundBoxEnabled, getBoundBoxPadding, getBoundBoxStrokeWidth } from './boundBoxUtils';
import { appendEle } from '../utils/zcsvgutils';
import { INNERLABEL_LABELTYPE, PARENTLABELS_LABELTYPE, DATALABELS_LABELTYPE } from './labeltype-constants';
import { polarToXY, isAngleInside } from '../polaraxis/polarutils';
import { d3_extent } from '../core/d3-constants';
import { mathSin, mathCos, mathSQRT, mathAtan2, mathAbs, mathTan } from '../core/zc-constants';
import { PIE_CHARTNAME } from '../core/chartname-constants';
import { translateTspan } from './tspan-utils';
import { isDataActive } from '../core/data-processor/is-data-active';
import { bubblepieUtils } from '../bubblepie/bubblepie-utils';
import { getChartTypeByIndex } from '../core/charttype-helpers';
import { isSeriesFiltered, hasColorColumnInNonAxis } from '../utils/colorutils';

function pieInnerLabel() {
    var renderer, args, chartObj, dataset;

    function labels(selection, labelHandler) {
        selection.each(function() {
            var currentdata = args[0],
                pieData = args[2],
                totalEngagement = args[4],
                translation = args[8] ? `translate(${args[8][0]},${args[8][1]})` : null,
                arc = args[7],
                isParent = args[9],
                rp = renderer.commonRendererProp().params,
                { rendererIndex = index } = rp;

            dataset = rp.dataset;
            chartObj = rp.chartObj;
            var commonPlotOptions = rp.commonPlotOptions,
                index = args[1] || 0,
                paneid = labelHandler.paneid,
                chartType = getChartTypeByIndex(chartObj, paneid, index).value,
                isSunburst = isSunburstChartType(chartType),
                outline = args[6],
                yaxiscolumnorder = rp.yaxiscolumnorder;

            isParent = isSunburst ? isParent : NULL;
            currentdata = currentdata[index];
            var isBubblePie = chartcategory.bubblePie.has(chartType);

            var plotoptions = currentdata.plotoptions,
                dataLabels = getMergedDatalabelOptions(commonPlotOptions, plotoptions, chartObj, index),
                innerLabels = dataLabels.innerLabel,
                showInnerLabel = innerLabels.show === true && datacount_based_intelligence(chartObj, DATALABELS_LABELTYPE);

            if (isParent) {
                var parentLabelOptions = mergeJSON(innerLabels.parentLabel, innerLabels);
                showInnerLabel = (defined(parentLabelOptions.show) ? parentLabelOptions.show : innerLabels.show) !== false;
                innerLabels = parentLabelOptions;
            }

            var GLT = dataLabelTextContent(chartObj, rp, innerLabels, INNERLABEL_LABELTYPE, totalEngagement, index, undefined, paneid),
                contentType = GLT.contentType || 'text',
                isMultiLine = contentType === 'html' && GLT.multiLine,
                textOverlap = innerLabels.textOverlap || 'adjust',
                boundBoxPadding = isBoundBoxEnabled(innerLabels, chartObj)
                    ? getBoundBoxPadding(innerLabels, chartObj) + getBoundBoxStrokeWidth(innerLabels, chartObj)
                    : 0;

            if (currentdata.disabled || !showInnerLabel || ((isBubblePie || isSunburst) && isSeriesFiltered(chartObj, paneid, index))) {
                return;
            }

            var temp_label = appendEle(chartObj.staticPaneSVG.select('.labelholder'), 'text', [1], '.', 'class', 'temp_label');
            applyDatalabelFonts(temp_label, innerLabels, chartObj, rp.color);
            temp_label.attr('transform', translation).style('opacity', 0);
            var queue = labelHandler.add;

            (isSunburst ? currentdata : hasColorColumnInNonAxis(chartObj) ? [currentdata] : currentdata.data).forEach(function(d, i) {
                var dataUpdate = isBubblePie ? (renderer.seriesdataAfterUpdate ? renderer.seriesdataAfterUpdate[i] : d) : pieData,
                    commonProperties = [],
                    subSeriesIndex = pick(args[10], i);

                dataUpdate.forEach(function(d, j) {
                    var bb = null;
                    var display = !isdisabled(d),
                        ANCHOR = 'middle',
                        DY = isMultiLine ? '0.9em' : '0.35em',
                        X,
                        Y,
                        xVal = 0,
                        yVal = 0;
                    if (isBubblePie) {
                        var stacked = rp.stacked,
                            axisrotated = rp.axisrotated,
                            xscale = rp.xscale,
                            yscale = rp.yscale;
                        var key = bubblepieUtils.getKey(d, index, chartObj);
                        var pieRenderer = chartObj.subrenderer[i][key];
                        d = pieRenderer.data[index];
                        arc = pieRenderer.arc;

                        xVal = chartObj.dataObject.getXYvalue(axisrotated, stacked, xscale, yscale, yaxiscolumnorder, d.data, 'x', null);
                        yVal = chartObj.dataObject.getXYvalue(axisrotated, stacked, xscale, yscale, yaxiscolumnorder, d.data, 'y', null);
                        var zVal = chartObj.dataObject.getXYvalue(
                            axisrotated,
                            stacked,
                            xscale,
                            yscale,
                            yaxiscolumnorder,
                            d.data,
                            'z',
                            chartObj.zscale
                        );

                        var outerRadius = zVal / 4;
                        var polarXY = polarToXY(outerRadius, midAngle(d));
                        GLT = dataLabelTextContent(
                            chartObj,
                            rp,
                            innerLabels,
                            INNERLABEL_LABELTYPE,
                            pieRenderer.total,
                            index,
                            undefined,
                            paneid
                        );

                        X = xVal + polarXY[0];
                        Y = yVal + polarXY[1];
                    } else {
                        X = arc.centroid(d)[0];
                        Y = arc.centroid(d)[1];
                    }

                    var dt = d.data;
                    var toNull = isSunburst ? d.depth == 0 : isdisabled(d);
                    var text = toNull ? null : GLT(dt, j);

                    temp_label[contentType](text).attrs({
                        x: X,
                        y: Y,
                        dy: DY,
                        'text-anchor': ANCHOR
                    });

                    if (display && isMultiLine) {
                        translateTspan(temp_label, X);
                        bb = bbox(temp_label.node());
                        Y -= bb.height / 2;
                    }

                    if (textOverlap != NONE) {
                        /*
                         * Default position for the inner label is arc.centroid(d)
                         * innerLabelWithinArc() finds the best position for the inner label
                         * If the inner label overflows in its centroid position but it could be placed inside the arc, it will be translated to its best position
                         * Else it will not be displayed
                         */
                        bb = bb || bbox(temp_label.node());
                        var centre = arc.centroid(d);
                        var wid = bb.width / 2 + boundBoxPadding,
                            hei = bb.height / 2 + boundBoxPadding;
                        var TL = {
                                x: centre[0] - wid,
                                y: centre[1] - hei
                            },
                            TR = {
                                x: centre[0] + wid,
                                y: centre[1] - hei
                            },
                            BL = {
                                x: centre[0] - wid,
                                y: centre[1] + hei
                            },
                            BR = {
                                x: centre[0] + wid,
                                y: centre[1] + hei
                            };
                        var isPointInArc = pointIsInArc(TL, d) && pointIsInArc(TR, d) && pointIsInArc(BL, d) && pointIsInArc(BR, d);
                        if (!isPointInArc) {
                            if (textOverlap === 'adjust') {
                                var SA = isSunburst ? outline.startAngle()(d) : d.startAngle,
                                    EA = isSunburst ? outline.endAngle()(d) : d.endAngle,
                                    MA = midAngle(d),
                                    extra = bb.height + 2 * boundBoxPadding, // / Math.cos(MA), //Need to check again
                                    s,
                                    e;

                                var a1 = xyToAngle(TL),
                                    a2 = xyToAngle(TR),
                                    a3 = xyToAngle(BL),
                                    a4 = xyToAngle(BR);
                                var angleArray = [a1, a2, a3, a4],
                                    posArray = [TL, TR, BL, BR];
                                var minmax = d3_extent(angleArray);

                                var s = posArray[angleArray.indexOf(minmax[0])],
                                    e = posArray[angleArray.indexOf(minmax[1])];

                                var dist = mathSQRT((s.x - e.x) * (s.x - e.x) + (s.y - e.y) * (s.y - e.y));
                                var r = mathAbs(dist / 2 / mathTan((EA - SA) / 2)) + extra;
                                var newCentre = [mathSin(MA) * r, -(mathCos(MA) * r)];

                                var TL = {
                                        x: newCentre[0] - wid,
                                        y: newCentre[1] - hei
                                    },
                                    TR = {
                                        x: newCentre[0] + wid,
                                        y: newCentre[1] - hei
                                    },
                                    BL = {
                                        x: newCentre[0] - wid,
                                        y: newCentre[1] + hei
                                    },
                                    BR = {
                                        x: newCentre[0] + wid,
                                        y: newCentre[1] + hei
                                    };
                                isPointInArc = pointIsInArc(TL, d) && pointIsInArc(TR, d) && pointIsInArc(BL, d) && pointIsInArc(BR, d);
                            }

                            if (isPointInArc) {
                                X = newCentre[0] + xVal;
                                Y = newCentre[1] + yVal;
                            } else {
                                display = false;
                            }
                        }
                    } else {
                        bb = bbox(temp_label.node());
                    }

                    var coordinate = {
                        X: X,
                        Y: Y,
                        DY: DY,
                        ANCHOR: ANCHOR,
                        text: text,
                        display: display,
                        ID: `innerlbl${index}_${subSeriesIndex}_${j}`,
                        width: bb.width,
                        height: bb.height
                    };
                    commonProperties.push(coordinate);
                });

                // add to drawing queue
                commonProperties.dataUpdate = dataUpdate;
                commonProperties.translation = translation;
                var q = queue(
                    commonProperties,
                    index,
                    subSeriesIndex,
                    NULL,
                    isParent ? PARENTLABELS_LABELTYPE : INNERLABEL_LABELTYPE,
                    rendererIndex,
                    paneid
                );
                q.properties = {
                    type: PIE_CHARTNAME,
                    dataLabels: innerLabels,
                    dataUpdate: dataUpdate,
                    showlabel: showInnerLabel,
                    rp: rp,
                    trans: args[8],
                    isSunburst: isSunburst,
                    contentType: contentType,
                    isMultiLine: isMultiLine,
                    chartType: chartType
                };
            });

            function pointIsInArc(pt, ptData) {
                var r1 = arc.innerRadius()(ptData),
                    r2 = arc.outerRadius()(ptData),
                    theta1 = arc.startAngle()(ptData),
                    theta2 = arc.endAngle()(ptData),
                    dist = pt.x * pt.x + pt.y * pt.y,
                    angle = xyToAngle(pt);
                return r1 * r1 <= dist && dist <= r2 * r2 && isAngleInside(theta1, theta2, angle);
            }

            function xyToAngle(obj) {
                return mathAtan2(obj.x, -obj.y);
            }

            function isdisabled(d) {
                var dataActive =
                    isDataActive(isSunburst ? d : d.data || d, chartObj) &&
                    (isBubblePie || isSunburst ? true : !isSeriesFiltered(chartObj, paneid, d.index));
                return !dataActive || (isSunburst ? d.parent == null || d.value == 0 : d.value == 0);
            }

            function midAngle(d) {
                var start = isSunburst ? outline.startAngle()(d) : d.startAngle,
                    end = isSunburst ? outline.endAngle()(d) : d.endAngle;
                return (start + end) / 2;
            }
        });
    }

    labels.renderer = function(_) {
        if (!arguments.length) {
            return renderer;
        }
        renderer = _;
        return labels;
    };
    labels.args = function(_) {
        if (!arguments.length) {
            return args;
        }
        args = _;
        return labels;
    };
    return labels;
}

LabelRegistry.add('pieInnerLabels', pieInnerLabel);
export var _pieInnerDatalabel = pieInnerLabel;
