// @zohocharts/zc/src/label/scatterdatalabels.js
//$Id$
//scatterdatalabels.js
/*
 * PENDING:
 * Bound box - minHeight should be considered for height on handling overlapping datalabels (if hook is enabled)
 */
import { LabelRegistry } from './label-registry';
import { charttype } from '../core/main-zc-init';
import { isScatterChartType } from '../events/events-utils';
import { AREARANGE_CHARTNAME, SCATTER_CHARTNAME } from '../core/chartname-constants';
import { getMergedDatalabelOptions, isDatalabelEnabled, applyDatalabelFonts, callTrimText, getLabelArray } from './datalabelsUtils';
import { getPlotBound } from '../plot/plotSize';
import {
    isBoundBoxEnabled,
    getMergedBoundBoxOptions,
    getBoundBoxHookConf,
    getBoundBoxPadding,
    getBoundBoxStrokeWidth
} from './boundBoxUtils';
import { mathMax, mathSQRT, mathPow, mathMin, mathPI } from '../core/zc-constants';
import { isMultiColoring, isSeriesFiltered } from '../utils/colorutils';
import { dataLabelTextContent } from './datalabel-content';
import { getFontSizeByHierarchy, getFontFamilyByHierarchy } from '../utils/zcdatautils';
import { appendEle } from '../utils/zcsvgutils';
import { NONE, NULL } from '../core/zc-constants';
import { isDataActive } from '../core/data-processor/is-data-active';
import { polarToXY, getDY } from '../polaraxis/polarutils';
import { DATALABELS_LABELTYPE } from './labeltype-constants';
import { translateTspan, trimAllTspan, getTspanBound } from './tspan-utils';
import { bbox, pick, simpleClone } from '../utils/zcutils';
import { getTextBound } from '../axis/textBound';
import { getSizeToBeAdded } from './boundBoxUtils';
import { BOTTOM_ALIGN } from '../core/align-constants';
import { bubblepieUtils } from '../bubblepie/bubblepie-utils';
import { d3_scaleLinear } from '../core/d3-constants';
import { DataProcessor } from '../core/data-processor';
import { chartcategory } from '../core/main-zc-init';
import { LOCATION_MARKERS } from '../geo-core/geo-marker-constant';
import { $Browser } from '../core/browser-detect';
import { getDy4MultiLine } from './position-helpers';
import { getChartTypeByIndex } from '../core/charttype-helpers';

function scatterLabel() {
    var renderer,
        args,
        labelseries,
        dataProcessorHelpers = DataProcessor.helpers;

    function labels(selection, labelHandler) {
        selection.each(function() {
            var currentdata = args[0],
                index = args[1],
                size = args[2] || 0,
                yrangePos = args[3],
                rendererIndex = pick(labelHandler.rendererIndex, index),
                paneid = labelHandler.paneid,
                transform = [0, 0],
                plotoptions = currentdata.plotoptions,
                dataLabels,
                rp = renderer.commonRendererProp().params,
                chartObj = rp.chartObj,
                systemconf = chartObj.systemConf,
                commonPlotOptions = rp.commonPlotOptions,
                dataset = rp.dataset,
                stacked = rp.stacked,
                axisrotated = rp.axisrotated,
                yaxiscolumnorder = rp.yaxiscolumnorder,
                xscale = rp.xscale,
                yscale = rp.yscale,
                labelArray = getLabelArray(rp, paneid),
                cType_Obj = getChartTypeByIndex(chartObj, paneid, index),
                cType = cType_Obj.value,
                isRangeChart = cType_Obj.name == AREARANGE_CHARTNAME,
                isCanvasMode = chartObj.plot.renderer.mode === 'canvas';
            var scatterOrBubble = isScatterChartType(cType),
                isBubbleChart = chartcategory.bubble.has(cType),
                isScatterChart = scatterOrBubble && !isBubbleChart,
                isBubblePie = chartcategory.bubblePie.has(cType),
                isPolar = chartObj.dataObject.isPolarAxisCategory,
                isGeoHeatmap = charttype.geoheatmap === cType,
                hasLocationMarker = isScatterChart && LOCATION_MARKERS.has(commonPlotOptions.marker.symbol);

            dataLabels = getMergedDatalabelOptions(commonPlotOptions, plotoptions, chartObj, index, paneid);
            var showlabel = isDatalabelEnabled(dataLabels, chartObj),
                labelOrient = pick(dataLabels.orient, 'top');

            if (currentdata.disabled || !showlabel || isSeriesFiltered(chartObj, paneid, index)) {
                return;
            }

            var currentPlotBound = getPlotBound(chartObj),
                boundBoxPadding = 0,
                boundBoxStrokeWidth = 0,
                hookHeight = 0,
                hookWidth = 0,
                boundBoxRadius = 0,
                plotareaBorderCorrection = getPlotareaBorderCorrection(dataLabels, chartObj);

            if (isBoundBoxEnabled(dataLabels, chartObj)) {
                var boundBoxOptions = getMergedBoundBoxOptions(dataLabels, chartObj);
                var hookOptions = getBoundBoxHookConf(boundBoxOptions.hook);
                boundBoxPadding = getBoundBoxPadding(dataLabels, chartObj);
                boundBoxStrokeWidth = getBoundBoxStrokeWidth(dataLabels, chartObj);
                hookHeight = hookOptions.height;
                hookWidth = hookOptions.width;
                boundBoxRadius = boundBoxOptions.radius;
            }

            var padding = isBubbleChart ? 0 : mathMax(0, +(dataLabels.padding || 0)) + boundBoxPadding + hookHeight,
                textOverlap = dataLabels.textOverlap, //hidden | adjust | ellipsis
                handleOverlapping = systemconf.chart.plot.datalabels.handleOverlapping === true,
                multiColoring = scatterOrBubble ? isMultiColoring(commonPlotOptions, chartObj) : false,
                left = currentPlotBound.left,
                top = currentPlotBound.top,
                right = currentPlotBound.right,
                bottom = currentPlotBound.bottom,
                leftWithBorderCorrection = left + plotareaBorderCorrection.left,
                topWithBorderCorrection = top + plotareaBorderCorrection.top,
                rightWithBorderCorrection = right - plotareaBorderCorrection.right,
                bottomWithBorderCorrection = bottom - plotareaBorderCorrection.bottom,
                GLT = dataLabelTextContent(chartObj, rp, dataLabels, DATALABELS_LABELTYPE, null, index, yrangePos, paneid),
                contentType = GLT.contentType || 'text',
                isMultiLine = contentType === 'html' && GLT.multiLine,
                isCatNullX = dataProcessorHelpers.isCatNull(dataset, 'x'),
                isCatNullY = dataProcessorHelpers.isCatNull(dataset, 'y', yaxiscolumnorder),
                fontSize = getFontSizeByHierarchy(dataLabels, chartObj),
                fontStyles = applyDatalabelFonts(null, dataLabels, chartObj, rp.color, false, index);

            var temp_label = appendEle(chartObj.staticPaneSVG.select('.labelholder'), 'text', [1], '.', 'class', 'temp_label');
            var updateHandler = labelHandler.update,
                queue = labelHandler.add,
                bottomOrient = labelOrient === 'bottom' && isScatterChart,
                factor = bottomOrient ? -1 : 1;

            currentdata.data.forEach(function(d, i) {
                var dataUpdate = renderer.seriesdataAfterUpdate ? renderer.seriesdataAfterUpdate[i] : currentdata.data[i];

                var calculated_points = [];
                applyDatalabelFonts(temp_label, dataLabels, chartObj, rp.color, false, index);
                temp_label.style('opacity', 0);

                dataUpdate.forEach(function(d, j) {
                    var xVal = chartObj.dataObject.getXYvalue(
                            axisrotated,
                            stacked,
                            xscale,
                            yscale,
                            yaxiscolumnorder,
                            d,
                            'x',
                            null,
                            yrangePos
                        ),
                        yVal = chartObj.dataObject.getXYvalue(
                            axisrotated,
                            stacked,
                            xscale,
                            yscale,
                            yaxiscolumnorder,
                            d,
                            'y',
                            null,
                            yrangePos
                        ),
                        zVal =
                            isBubbleChart || isBubblePie
                                ? chartObj.dataObject.getXYvalue(
                                      axisrotated,
                                      stacked,
                                      xscale,
                                      yscale,
                                      yaxiscolumnorder,
                                      d,
                                      'z',
                                      chartObj.zscale
                                  )
                                : null,
                        isWithinClrRange = isDataActive(d, chartObj);
                    var x = null,
                        y = null,
                        text = null,
                        tspanInfo = null,
                        display = isWithinClrRange && !d.disabled,
                        id = '',
                        dy = isMultiLine ? '0.9em' : isBubbleChart || isGeoHeatmap || bottomOrient ? '0.3em' : '-0.3em',
                        ANCHOR = 'middle',
                        HOOKPOSITION = isBubbleChart ? null : BOTTOM_ALIGN,
                        xDiff = 0,
                        yDiff = 0,
                        midAngle,
                        polarXY;

                    if (isBubblePie) {
                        var key = bubblepieUtils.getKey(d, index, chartObj);
                        var pieRenderer = chartObj.subrenderer[i][key];
                        var pieData = pieRenderer.data[index];
                        midAngle = (pieData.startAngle + pieData.endAngle) / 2;
                        var outerRadius = zVal / 2 + padding;
                        var polarXY = polarToXY(outerRadius, midAngle);
                        dy = isMultiLine ? '0.9em' : midAngle > mathPI / 2 && midAngle < (3 * mathPI) / 2 ? `${getDY(midAngle)}em` : '0';
                        ANCHOR = midAngle < 0 || midAngle > mathPI ? 'end' : 'start';
                        GLT = dataLabelTextContent(chartObj, rp, dataLabels, DATALABELS_LABELTYPE, pieRenderer.total, index, yrangePos);
                    }
                    temp_label[contentType](function() {
                        if (d.isDummy || isCatNullX(dataset.getX(d)) || isCatNullY(dataset.getY(d, yaxiscolumnorder, yrangePos))) {
                            text = null;
                        } else {
                            text = GLT(d, j);
                        }
                        return text;
                    })
                        .styles({
                            display: function() {
                                if (isPolar) {
                                    return null;
                                }
                                //If the data point itself out of plot, need not to show label.
                                var ptX = xVal,
                                    ptY = yVal,
                                    pointOutOfPlot = ptX < left || ptX > right || ptY < top || ptY > bottom;
                                return pointOutOfPlot ? ((display = false), NONE) : null;
                            }
                        })
                        .attrs({
                            transform: null,
                            dy: dy,
                            x: function() {
                                x = isBubblePie ? xVal + polarXY[0] : xVal;
                                return x || 0;
                            },
                            y: function() {
                                if (hasLocationMarker) {
                                    if (bottomOrient) {
                                        radius = 2;
                                    } else {
                                        radius = size;
                                    }
                                } else {
                                    radius = (isBubbleChart ? 0 : size) / 2;
                                }

                                radius *= factor;

                                y = isBubblePie
                                    ? yVal + polarXY[1]
                                    : yVal + (radius + padding * factor) * (isRangeChart && !yrangePos ? 1 : -1);
                                return y || 0;
                            },
                            'text-anchor': ANCHOR
                        });

                    // ID
                    id = `lbl${rendererIndex}_${i}_${j}_${index}_${yrangePos}`;

                    var textBound;
                    if (isMultiLine) {
                        let t = getDy4MultiLine(temp_label, dataLabels, parseFloat(dy), text, display);
                        text = t[2];
                        display = t[3];

                        translateTspan(temp_label, x);
                        textBound = bbox(temp_label.node());

                        if (isBubblePie) {
                            var yAdjust = d3_scaleLinear()
                                .range([-1, 0])
                                .domain(midAngle < mathPI ? [0, mathPI] : [2 * mathPI, mathPI]);
                            y += yAdjust(midAngle) * textBound.height;
                            translateTspan(temp_label, x);
                        } else {
                            if (!bottomOrient) {
                                // top orient
                                //#ZC1717
                                y -= isBubbleChart || isGeoHeatmap ? textBound.height / 2 : textBound.height;
                            }
                        }

                        temp_label.attr('y', y);
                        textBound = bbox(temp_label.node());
                    } else if (contentType === 'html') {
                        //Test_665.json
                        textBound = bbox(temp_label.node());
                    } else {
                        textBound = getTextBound(
                            text,
                            fontSize,
                            getFontFamilyByHierarchy(dataLabels, chartObj),
                            null,
                            null,
                            fontStyles['font-style'],
                            fontStyles['font-weight']
                        );
                        textBound.x = x - (ANCHOR === 'middle' ? textBound.width / 2 : ANCHOR === 'end' ? textBound.width : 0);
                        textBound.y = y - textBound.height;
                        // #ZC1727
                        if (bottomOrient) {
                            y += textBound.height / 2;
                        }
                    }

                    // Adjust Text Within Bound
                    var widthToBeAdded = 0;
                    if (!isPolar) {
                        widthToBeAdded = getSizeToBeAdded(textBound.width, boundBoxPadding, boundBoxRadius, hookWidth);
                        var adjust = boundBoxPadding + boundBoxStrokeWidth / 2 + widthToBeAdded / 2,
                            scrolledX = 0; //chartObj.chartRendered ? (chartObj.scrolledX || 0) : 0;

                        var label = temp_label,
                            leftPos = textBound.x - adjust,
                            rightPos = leftPos + textBound.width + scrolledX + 2 * adjust;
                        var newX =
                            x +
                            (leftPos < leftWithBorderCorrection
                                ? leftWithBorderCorrection - leftPos
                                : rightPos > rightWithBorderCorrection
                                ? -(rightPos - rightWithBorderCorrection)
                                : 0);

                        if ((xDiff = newX - x)) {
                            x = newX;
                            label.attr('x', newX);
                            if (isMultiLine) {
                                translateTspan(temp_label, newX);
                            }
                        }

                        // var z = +label.attr("y");
                        var newY = y + (textBound.y < topWithBorderCorrection ? textBound.height + size + padding : 0);
                        if ((yDiff = newY - y)) {
                            // if (chartObj.dataObject.isPercentChart && !axisrotated) {
                            y = newY;
                            label.attr('y', newY);
                            // } else {
                            //     display = false;
                            // }
                        }

                        if (isMultiLine && (xDiff || yDiff)) {
                            textBound = bbox(temp_label.node());
                        } else {
                            if ($Browser.IE) {
                                // #ZC1751
                                // IE does not allow to the SVGRect object
                                textBound = simpleClone(textBound);
                            }
                            textBound.x += xDiff;
                            textBound.y += yDiff;
                        }
                    }

                    widthToBeAdded = isPolar
                        ? getSizeToBeAdded(textBound.width, boundBoxPadding, boundBoxRadius, hookWidth)
                        : widthToBeAdded;
                    var bb_width = textBound.width + widthToBeAdded,
                        bb_height = textBound.height,
                        bb_x = textBound.x + widthToBeAdded / 2;
                    if (handleOverlapping) {
                        var radius = (isBubbleChart ? zVal : size) / 2;

                        if (isBubbleChart) {
                            // Bubble chart - Don't show datalabels when they are out of the bubble bound
                            var cx = xVal,
                                cy = yVal,
                                leftDist = mathSQRT(mathPow(cx - bb_x, 2) + mathPow(cy - textBound.y, 2)) + boundBoxPadding,
                                rightDist =
                                    mathSQRT(
                                        mathPow(cx - (bb_x + textBound.width), 2) + mathPow(cy - (textBound.y + textBound.height), 2)
                                    ) + boundBoxPadding;

                            if (leftDist > radius || rightDist > radius) {
                                var trimmed;
                                if (textOverlap === 'adjust') {
                                    dy = isMultiLine ? '0.9em' : '-0.3em';
                                    HOOKPOSITION = BOTTOM_ALIGN;
                                    var radius = zVal / 2;
                                    radius *= factor;
                                    var _padding = mathMax(0, +(dataLabels.padding || 0)) + boundBoxPadding + hookHeight;
                                    y = yVal + (radius + _padding * factor) * (isRangeChart && !yrangePos ? 1 : -1);
                                    y = y || 0;
                                    textBound.y = y - textBound.height;

                                    // ZC3169
                                    if (textBound.y < topWithBorderCorrection) {
                                        y = textBound.height;
                                        textBound.y = y;
                                    }
                                } else {
                                    if (textOverlap === 'ellipsis') {
                                        // add ellipsis
                                        var h = mathMin(textBound.height / 2, radius);
                                        var w = mathSQRT(mathPow(radius, 2) - mathPow(h, 2));
                                        var bubbleBound = {
                                            left: cx - w + boundBoxPadding,
                                            top: cy - w + boundBoxPadding,
                                            right: cx + w - boundBoxPadding,
                                            bottom: cy + w - boundBoxPadding
                                        };

                                        if (contentType === 'html') {
                                            tspanInfo = getTspanBound(temp_label, fontSize);
                                            trimAllTspan(temp_label, chartObj, transform, 0, bubbleBound, tspanInfo);
                                        } else {
                                            callTrimText(temp_label, chartObj, transform, 0, bubbleBound);
                                        }

                                        var trimmedText = temp_label.text();
                                        if (trimmedText && trimmedText !== '..') {
                                            //If the textContent is just "..", hide the label
                                            text = temp_label[contentType]();
                                            trimmed = true;
                                            bb_width = 2 * w - 2 * boundBoxPadding;
                                            if (isMultiLine) {
                                                var newBB = bbox(temp_label.node());
                                                bb_height = newBB.height;
                                                bb_width = newBB.width + widthToBeAdded;
                                                y += (textBound.height - newBB.height) / 2;
                                            }
                                        }
                                    }
                                    if (!trimmed) {
                                        display = false;
                                        temp_label.style('display', NONE);
                                    }
                                }
                            }
                        }

                        // only render if the label is visible
                        // and it should be implemented in all charts
                        // TODO Remila need to check OLD ZR 282
                        if ((chartObj.isGeoMap && display) || !chartObj.isGeoMap) {
                            var obj = {
                                id: id,
                                x1: bb_x - boundBoxPadding,
                                y1: textBound.y - boundBoxPadding,
                                x2: bb_x + bb_width + boundBoxPadding,
                                y2: textBound.y + textBound.height + boundBoxPadding + hookHeight,
                                pointRadius: radius + padding - hookHeight,
                                translateFunction: updateHandler,
                                overlap: textOverlap,
                                hookHeight: hookHeight,
                                orient: labelOrient,
                                factor: factor
                            };
                            labelArray.push(obj);
                        }
                    }

                    var coordinate = {
                        X: x,
                        Y: y,
                        DY: dy,
                        ANCHOR: ANCHOR,
                        text: text,
                        display: display,
                        height: bb_height,
                        width: bb_width,
                        hookPosition: HOOKPOSITION,
                        adjustVal: xDiff
                    };

                    if (isCanvasMode && contentType === 'html') {
                        coordinate.tspanInfo = tspanInfo || getTspanBound(temp_label, fontSize);
                    }

                    calculated_points.push(coordinate);
                });

                calculated_points.dataUpdate = dataUpdate; //Each subseries may have different dataUpdates
                var q = queue(calculated_points, index, i, yrangePos, NULL, rendererIndex);
                q.properties = {
                    type: SCATTER_CHARTNAME,
                    dataLabels: dataLabels,
                    color: rp.color,
                    dataUpdate: dataUpdate,
                    showlabel: showlabel,
                    multiColoring: multiColoring,
                    rp: rp,
                    rangeChart: isRangeChart,
                    yrangePos: yrangePos,
                    contentType: contentType,
                    isMultiLine: isMultiLine,
                    chartType: cType
                };
            });
        });
    }

    function getPlotareaBorderCorrection(dataLabels, chartObj) {
        var plotareaBorderEle = chartObj.staticPaneSVG.selectAll('g.plotareaBorder-property').node(),
            defaultVal = 1,
            correction = {
                left: 0,
                top: 0,
                right: 0,
                bottom: 0
            };

        if (isBoundBoxEnabled(dataLabels, chartObj)) {
            for (var side in correction) {
                correction[side] = plotareaBorderEle[`border-${side}`]
                    ? plotareaBorderEle[`border-${side}`].size / 2 + defaultVal
                    : defaultVal;
            }
        }

        return correction;
    }

    labels.renderer = function(_) {
        if (!arguments.length) {
            return renderer;
        }
        renderer = _;
        return labels;
    };
    labels.args = function(_) {
        if (!arguments.length) {
            return args;
        }
        args = _;
        return labels;
    };

    return labels;
}

LabelRegistry.add('scatter', scatterLabel);
export var _scatterDatalabel = scatterLabel;
