// @zohocharts/zc/src/label/targetlabels.js
//$Id$
//targetlabels.js
import { LabelRegistry } from './label-registry';
import { getPlotBound } from '../plot/plotSize';
import { appendEle, applyFont } from '../utils/zcsvgutils';
import { BOTTOM_ALIGN, LEFT_ALIGN, CENTER_ALIGN, TOP_ALIGN, RIGHT_ALIGN } from '../core/align-constants';
import { defined, validateNullVal } from '../utils/zcutils';
import { getLabelArray } from './datalabelsUtils';
import { getFormattedValue } from '../formatter/formatter';
import { d3_select, d3_ascending } from '../core/d3-constants';
import { $ZCG } from '../core/zcglobal';

function targetLabel(targetMarkerOptions, renderer, dataUpdate, options) {
    var targetLabelOptions = targetMarkerOptions.label || {};
    var chartObj = this,
        dataset = this.dataset,
        plotarea = this.plotarea,
        systemconf = this.systemConf,
        commonOption = renderer.commonOption,
        rp = renderer.commonRendererProp().params,
        axisrotated = rp.axisrotated,
        yscale = rp.yscale,
        rangeBand = rp.xscale.bandwidth(),
        yaxisOrient = systemconf.chart.axes.yaxis[0].orient || (this.axes.rotated ? BOTTOM_ALIGN : LEFT_ALIGN),
        labelArray = getLabelArray(rp, options.paneid);

    var targetDataIndex = targetMarkerOptions.dataindex,
        showlabel = targetMarkerOptions.enabled === true && targetLabelOptions.enabled === true && defined(targetDataIndex);

    var wrap = this.staticPaneSVG.selectAll('g.qualitativeGroup'),
        targetLabelGroup = appendEle(wrap, 'g', showlabel ? [1] : [], '.', 'class', 'targetLabel');
    targetLabelGroup.attr('transform', `translate(${plotarea.left},${plotarea.top})`);

    if (!showlabel) {
        return;
    }

    var hAlign = targetLabelOptions.hAlign,
        vAlign = targetLabelOptions.vAlign,
        { textOverlap = 'none' } = targetLabelOptions,
        padding = validateNullVal(targetLabelOptions.padding, 2),
        isFixedPosition = targetLabelOptions.position === 'fixed',
        markerSize =
            validateNullVal(
                targetMarkerOptions.size,
                (commonOption.levelMarkerWidth - 2 * validateNullVal(targetMarkerOptions.padding, commonOption.padding / 2)) / 10
            ) * 10,
        markerSizeAdjust =
            (targetMarkerOptions.symbol === 'line' ? validateNullVal(targetMarkerOptions.outerStrokeWidth, 1) : markerSize) / 2;

    if (axisrotated) {
        hAlign = hAlign || CENTER_ALIGN;
        vAlign = vAlign || (yaxisOrient === BOTTOM_ALIGN ? TOP_ALIGN : BOTTOM_ALIGN);
    } else {
        hAlign = hAlign || (yaxisOrient === LEFT_ALIGN ? RIGHT_ALIGN : LEFT_ALIGN);
        vAlign = vAlign || CENTER_ALIGN;
    }

    var columns = this.userdata.metadata.columns;
    var columnidx = columns.filter(function(d) {
        return d.dataindex === targetMarkerOptions.dataindex;
    });
    var targetColumn = columnidx.length ? columnidx[0] : columns[dataset.getColumnIdx('y', 0)];
    var datatype = targetColumn.datatype;
    var formatConf = targetColumn[datatype];

    var adjust = {
        left: 0,
        right: 0,
        top: 0,
        bottom: 0
    };
    if (axisrotated) {
        adjust.top = adjust.bottom = plotarea.adjust;
    } else {
        adjust.left = adjust.right = plotarea.adjust;
    }
    if (!isFixedPosition && systemconf.chart.axes.yaxis[0].show === true) {
        adjust[yaxisOrient] = 0;
    }

    var rectBound = chartObj.container.node().getBoundingClientRect(),
        rectBoundLeft = rectBound.left + chartObj.plotarea.left,
        rectBoundTop = rectBound.top + chartObj.plotarea.top,
        currentPlotBound = getPlotBound(chartObj),
        left = currentPlotBound.left - adjust.left,
        top = currentPlotBound.top - adjust.top,
        right = currentPlotBound.right + adjust.left + adjust.right,
        bottom = currentPlotBound.bottom + adjust.top + adjust.bottom;

    applyFont(targetLabelGroup, [targetLabelOptions, systemconf.chart, systemconf.canvas, $ZCG]);
    var dataKeyFunc = function(d) {
        return dataset.getX(d);
    };
    var targetLabelText = appendEle(targetLabelGroup, 'text', dataUpdate, null, null, null, null, dataKeyFunc);
    targetLabelText.style('display', null);

    targetLabelText.each(function(d, i) {
        var targetValue = d[targetDataIndex];
        if (!defined(targetValue)) {
            return;
        }

        var xValue = dataset.getX(d),
            X,
            Y,
            DY,
            ANCHOR,
            TEXT = defined(targetLabelOptions.text)
                ? dataset.getCustomizedVal(chartObj, targetLabelOptions.text, 0, null, null, d, 'targetLabel')
                : getFormattedValue(formatConf, targetValue, datatype, chartObj, 'targetLabel'),
            isNeg = targetValue < 0,
            flip = isNeg ? -1 : 1;
        var x = commonOption.bandwidth1(xValue) + rangeBand / 2,
            y = yscale(targetValue);

        if (axisrotated) {
            X = y;
            Y = x;
            if (vAlign === BOTTOM_ALIGN) {
                Y += markerSize / 2;
                if (hAlign === CENTER_ALIGN) {
                    DY = '1em';
                    Y = x + (isFixedPosition ? rangeBand : markerSize) / 2;
                }
            } else if (vAlign === CENTER_ALIGN) {
                DY = '0.3em';
            } else if (vAlign === TOP_ALIGN) {
                Y -= markerSize / 2;
                if (hAlign === CENTER_ALIGN) {
                    DY = '-0.3em';
                    Y = x - (isFixedPosition ? rangeBand : markerSize) / 2;
                } else {
                    DY = '1em';
                }
            }

            if (hAlign === CENTER_ALIGN) {
                ANCHOR = 'middle';
                Y -= vAlign === CENTER_ALIGN ? 0 : vAlign === TOP_ALIGN ? padding * flip : -(padding * flip);
            } else {
                if (hAlign === RIGHT_ALIGN) {
                    X += (padding + markerSizeAdjust) * flip;
                    ANCHOR = isNeg ? 'end' : null;
                } else {
                    X -= (padding + markerSizeAdjust) * flip;
                    ANCHOR = isNeg ? null : 'end';
                }
            }
        } else {
            X = x;
            Y = y;
            if (hAlign === RIGHT_ALIGN) {
                X += (isFixedPosition ? rangeBand : markerSize) / 2;
            } else if (hAlign === CENTER_ALIGN) {
                ANCHOR = 'middle';
            } else if (hAlign === LEFT_ALIGN) {
                X -= (isFixedPosition ? rangeBand : markerSize) / 2;
                ANCHOR = 'end';
            }

            if (vAlign === CENTER_ALIGN) {
                X += hAlign === CENTER_ALIGN ? 0 : hAlign === RIGHT_ALIGN ? padding : -padding;
                DY = '0.3em';
            } else if (vAlign === TOP_ALIGN) {
                Y -= (padding + markerSizeAdjust) * flip;
                DY = '-0.3em';
            } else {
                Y += (padding + markerSizeAdjust) * flip;
                DY = '1em';
            }
        }

        d3_select(this)
            .attrs({
                transform: `translate(${X},${Y})`,
                'text-anchor': ANCHOR,
                dy: DY
            })
            .text(TEXT);

        // label should be within container
        var b = this.getBoundingClientRect(),
            leftPos = b.left - rectBoundLeft,
            rightPos = leftPos + b.width,
            topPos = b.top - rectBoundTop,
            bottomPos = topPos + b.height;

        if (textOverlap !== 'none') {
            var ele = d3_select(this),
                obj = {
                    id: `lbl${options.rendererIndex}_0_${options.subIndex}_${i}`,
                    translateFunction: function(id, translateX, translateY, hide) {
                        if (hide) {
                            ele.style('display', 'none');
                        } else {
                            ele.attr('transform', `translate(${X + translateX},${Y + translateY})`);
                        }
                    },
                    x1: leftPos,
                    y1: topPos,
                    x2: rightPos,
                    y2: bottomPos,
                    overlap: textOverlap,
                    priority: 1
                };
            labelArray.push(obj);
            labelArray.accessor = (a, b) => d3_ascending(a.priority || 0, b.priority || 0);
            labelArray.align = 'vertical';
        }

        var newX = X + (leftPos < left ? left - leftPos : rightPos > right ? -(rightPos - right) : 0),
            newY = Y + (topPos < top ? top - topPos : bottomPos > bottom ? -(bottomPos - bottom) : 0);

        if (newX != X) {
            X = newX;
        }

        if (newY != Y) {
            Y = newY;
        }

        d3_select(this).attr('transform', `translate(${X},${Y})`);
    });
}

LabelRegistry.add('bullet', targetLabel);
export var _bulletDatalabel = targetLabel;
