// @zohocharts/zc/src/label/tspan-utils.js
//$Id$
//tspan-utils.js
import { $Browser } from '../core/browser-detect';
import { isArray, bbox } from '../utils/zcutils';
import { BOTTOM_ALIGN } from '../core/align-constants';
import { TSPAN_ELEMENT, TEXT_ELEMENT } from '../core/element-constants';
import { d3_select } from '../core/d3-constants';
import { trimText } from '../utils/zcsvgutils';
import { getTspanBoundingClientRect } from '../axis/textBound';

export function getTspanTemplate(showAs) {
    var newShowAs = '';
    if (isArray(showAs)) {
        showAs.forEach(function(d, i) {
            newShowAs += `<tspan class='zcTspan'${i ? " dy='1.1em'" : ''}>${updateTspanTemplate(d)}</tspan>`;
        });
    } else {
        var tagIndex = showAs.indexOf('<tspan');
        if (tagIndex != -1) {
            newShowAs = `<tspan class='zcTspan'>${updateTspanTemplate(showAs)}</tspan>`;
        } else {
            newShowAs = showAs;
        }
    }
    return newShowAs;
}

function updateTspanTemplate(showAs) {
    var str = '',
        lastIndex = 0,
        endTagLength = 8,
        t = {
            '<tspan': [],
            '</tspan>': []
        };

    showAs.replace(/<tspan|<\/tspan>/g, function() {
        t[arguments[0]].push(arguments[1]);
    });

    t['<tspan'].forEach(function(d, i) {
        if (lastIndex != d) {
            str += `<tspan>${showAs.substring(lastIndex, d)}</tspan>`;
        }
        str += showAs.substring(d, (lastIndex = t['</tspan>'][i] + endTagLength));
    });

    if (lastIndex != showAs.length) {
        str += `<tspan>${showAs.substring(lastIndex, showAs.length)}</tspan>`;
    }

    return str;
}

export function translateTspan(ele, x) {
    ele.selectAll('tspan.zcTspan').attr('x', x);
}

export function trimAllTspan(labelText, chartObj, tranform, fontSize, edge, b, boundboxPadding, xy, preCalculation = {}) {
    if (!b.length) {
        // ZC3001 Error: before chart rendering, tab switching may happen.
        return;
    }

    var dx = +tranform[0],
        dy = +tranform[1],
        plotarea = chartObj.plotarea,
        margin = {
            left: -fontSize,
            right: -fontSize,
            top: 0,
            bottom: 0
        },
        pattern = '..',
        orient = BOTTOM_ALIGN,
        { useBBox } = preCalculation;
    boundboxPadding = boundboxPadding || 0;

    edge = edge || {
        left: -dx + boundboxPadding,
        top: -dy,
        right: plotarea.width - dx - boundboxPadding,
        bottom: plotarea.height - dy
    };

    labelText.each(function() {
        var children = this.childNodes,
            childCount = 0,
            parentWidth;

        if (!useBBox) {
            parentWidth = getParentWidth(this, edge, boundboxPadding, preCalculation);
        }

        for (var i = 0; i < children.length; i++) {
            if (useBBox) {
                parentWidth = getParentWidth(this, edge, boundboxPadding, preCalculation);
            }

            var node = children[i],
                c = node.childNodes,
                totalWidth = node.getComputedTextLength(),
                textContent = '';

            //collect all text contents
            for (var j = 0; j < c.length; j++) {
                var curTextContent = c[j].textContent;
                textContent += curTextContent;
                b[childCount + j].textLength = curTextContent.length;
            }

            // trim the text
            if (totalWidth - parentWidth > 1) {
                //#ZC1113
                var bb = b[childCount].bbox;
                var textBound = {
                    left: bb.x + (xy ? xy[0] : 0),
                    top: bb.y + (xy ? xy[1] : 0),
                    width: totalWidth,
                    height: bb.height
                };
                trimText(c[0], textContent, textBound, edge, orient, pattern, margin, null, null, false);

                // distribute the text contents to the respective tspan
                var innerHTML = c[0].innerHTML;
                const text_length = textContent.length, 
                    pattern_length = pattern.length;
                (innerHTML.length > text_length) ? (innerHTML = innerHTML.substring(0, text_length - pattern_length) + pattern) : null;
                if (innerHTML === '..') {
                    innerHTML = '';
                }

                if (useBBox) {
                    let newB = getTspanBound(d3_select(this), fontSize, useBBox);
                    b.forEach((_b, i) => (newB[i].textLength = _b.textLength));
                    b = newB;
                }

                for (var j = 0, start = 0; j < c.length; j++) {
                    var curTextLength = b[childCount + j].textLength;
                    c[j].innerHTML = innerHTML.substring(start, start + curTextLength);
                    start += curTextLength;
                }
            }
            childCount += useBBox ? c.length : c.length - 1;
        }
    });
}

function getParentWidth(node, edge, boundboxPadding, preCalculation) {
    var parentBBox = bbox(node),
        parentWidth = preCalculation.useBBox ? parentBBox.width : node.getComputedTextLength();

    if (parentBBox.x + parentWidth > edge.right) {
        parentWidth -= parentBBox.x + parentWidth - edge.right + boundboxPadding;
    }
    if (parentBBox.x < edge.left) {
        parentWidth -= edge.left - parentBBox.x + boundboxPadding;
    }
    return parentWidth;
}

export function getTspanBound(label, fontSize, useBBox) {
    var c = label.selectAll(function() {
            return this.childNodes;
        }),
        anchor = label.attr('text-anchor'),
        X = +label.attr('x'),
        Y = +label.attr('y'),
        o = [],
        DY = 0;

    c.each(function(_, m) {
        if (!isVisibleTspan(this)) {
            if (useBBox) {
                o.push({});
            }

            return;
        }

        var ele = d3_select(this),
            nodeDY = parseFloat(ele.attr('dy') || 0),
            children = this.childNodes,
            totalWidth = this.getComputedTextLength(),
            oldRef = anchor === 'middle' ? -totalWidth / 2 : anchor == 'end' ? -totalWidth : 0;
        DY += nodeDY;

        for (var i = 0; i < children.length; i++) {
            var currentWidth = children[i].getComputedTextLength();
            var info = {
                dy: DY,
                dx: oldRef + (anchor === 'middle' ? currentWidth / 2 : 0),
                bbox: { x: X + oldRef, y: Y + DY * fontSize, width: children[i].getComputedTextLength(), height: fontSize },
                style: children[i].style,
                text: children[i].textContent,
                id: m + '_' + i
            };
            o.push(info);
            oldRef += currentWidth;
        }
    });
    return o;
}

export function getCustomHTML(element, pool) {
    pool = pool || [];
    element.each(function(d, i) {
        var children = d3_select(this).selectAll('*');
        if (children.size()) {
            children.each(function() {
                getCustomHTML(d3_select(this), pool);
            });
        } else {
            var bb = bbox(this),
                nodeName = this.nodeName;
            if (nodeName === TSPAN_ELEMENT) {
                if (this.textContent === '') {
                    //If tspan is trimmed and text content is nothing, getTspanBoundingClientRect(this) will return null
                    return;
                }

                var parentNode = findParentTextNode(this),
                    parentBCR = getTspanBoundingClientRect(parentNode),
                    bcr = getTspanBoundingClientRect(this);
                bb.x += bcr.left - parentBCR.left;
                bb.y += bcr.top - parentBCR.top;
                bb.width = bcr.width;
                bb.height = bcr.height;
            }
            pool.push({
                node: this,
                nodeName: nodeName,
                bb: bb
            });
        }
    });
    return pool;
}

function findParentTextNode(node) {
    if (node.parentNode.nodeName !== TEXT_ELEMENT) {
        return findParentTextNode(node.parentNode);
    }
    return node.parentNode;
}

/**
 * Return all visible tspan
 * @param  {HTMLElement} ele
 * @return {Array}     Array of HTMLElement
 */
export function getVisibleTspan(ele) {
    let nodes = ele.selectAll('tspan.zcTspan').nodes();

    return Array.prototype.filter.call(nodes, isVisibleTspan);
}

/**
 * Return whether the tspan is visible or not
 * @param  {HTMLElement} ele
 * @return {Boolean}
 */
function isVisibleTspan(node) {
    let minSize = $Browser.PHANTOMJS || $Browser.IE ? 1 : 0,
        bb = getTspanBoundingClientRect(node);

    return bb.width > minSize && bb.height > minSize; //why 1: in phantomjs eventhough tspan is empty getTspanBoundingClientRect returns {width: 1, height: 1} instead of {width: 0, height: 0}  FEATURES_DATALABEL_10
}
