// @zohocharts/zc/src/legend/size-legend/list/expandable-helper.js
//
import { HORIZONTAL_ALIGN } from '../../../core/align-constants';
import { ListLegend } from '../../list-legend/list-legend';
import { mathMax, mathRound } from '../../../core/zc-constants';
import { d3_selectAll } from '../../../core/d3-constants';
import { createElement } from '../../../utils/zcsvgutils';
import { defined, getBoundingRect, isUndefined } from '../../../utils/zcutils';
import { getLegendAlignment } from '../../legend-utils';
import { legend_helpers } from '../../../core/main-zc-init';
import { SERIES_LEGEND_CLASSNAME } from '../../legend-constants';

export function getVisibleItemRenderCount(instance) {
    let spec = instance.spec,
        chart = spec.chartObj;
    let { config, bound } = spec,
        { border: borderConfig, title: titleConfig, list: listConfig } = config,
        { expandable: expandableConfig = {} } = listConfig,
        borderWidth = borderConfig && borderConfig.show ? borderConfig.size : 0,
        { maxWidth, maxHeight } = bound,
        align = getLegendAlignment(config),
        isHorizontal = align.layout === HORIZONTAL_ALIGN,
        isSizeLegend = spec.encode === 'z',
        legData = !isSizeLegend && chart ? legend_helpers.getLegendData(chart, spec.order) : spec.data,
        legData_length = !isSizeLegend && chart ? legend_helpers.getLegendDataLength(chart, legData) : spec.data.length;

    if (!expandableConfig.show) {
        return;
    }
    if (isHorizontal && !isSizeLegend) {
        let extraSpaces = +config.marginLeft + +config.marginRight, // + (config.border.show ? 2 * config.border.size : 0),
            allowedLegMaxWidth = chart.canvasarea.width - extraSpaces;

        if (maxWidth > allowedLegMaxWidth) {
            maxWidth = allowedLegMaxWidth;
        }
    }

    let // Temp element
        tempEle = createElement('div', '.temp_list_view', [1], d3_selectAll('div.__baseZC_Container__')),
        maxW = mathRound(maxWidth - (isSizeLegend ? 2 * borderWidth : 0)),
        maxH = mathRound(maxHeight - (isSizeLegend ? 2 * borderWidth : 0)), //we removed borderWidth here since below in 'getBoundingRect', we will get along with border only, so comparison wont be proper if we neglect border here.
        i = 0,
        w = 0,
        h = 0,
        actualTitleShow = titleConfig.show,
        actualExpanded = expandableConfig._expanded;

    //the maxWidth and maxHeight are roundedoff, since in some cases if maxWidth is 200.5, but in boundbox calculation we will get roundedoff values only, so inorder make calculations proper we round off them.
    tempEle.styles({
        display: 'block',
        position: 'fixed'
    });
    expandableConfig._expanded = false; // ZC2487
    // if (isHorizontal && titleConfig.keepAligned) {
    //     // If title is exceeds max-width, visibleItemCount is 1 always. So disabling title

    //     //reason: when title considered, the boundbox height got should not include the title's height but its including, so to avoid that this isWithinRange is made
    //     titleConfig.show = false;
    // }
    let _viewSpec = { ...spec, data: chart ? legend_helpers.getLegendData(chart, spec.order) : instance.viewScaleComponent().ticks },
        _viewObj = new ListLegend(_viewSpec, instance.viewScaleComponent);

    _viewSpec.isPreRender = true;

    var temp_i = i,
        itemCountInFirstrow,
        prev_h = 0,
        prev_w = 0,
        series_height,
        offsetTop,
        offsetLeft,
        series_width,
        isWidthInLimit,
        isHeightInLimit,
        temp_h,
        temp_w,
        isWithinRange,
        title_display,
        rowCount = 0,
        dummy,
        currentRow_rendered = false,
        nonHiddenItemsCount = 0,
        expandable_height,
        prevH_without_expandable;

    while ((isWithinRange || !currentRow_rendered) && ++temp_i <= spec.data.length) {
        if (spec.data[temp_i - 1].hideInLegend) {
            continue;
        }
        // nonHiddenItemsCount++;
        _viewSpec.visibleItemCount = temp_i;
        _viewSpec.nonHidden_visibleItemCount = ++nonHiddenItemsCount;

        _viewObj.create();
        _viewObj.draw(tempEle);
        let b = getBoundingRect(tempEle);

        if (isHorizontal) {
            w = mathRound(b.width);
            h = mathRound(b.height);
        } else {
            h = mathRound(b.height);
        }

        if (!isSizeLegend) {
            if (titleConfig.keepAligned && isHorizontal) {
                title_display = tempEle.selectAll('.zc-header-legseries').style('display');
                tempEle.selectAll('.zc-header-legseries').style('display', 'none');
                h = tempEle.node().offsetHeight;
                tempEle.selectAll('.zc-header-legseries').style('display', title_display);
            }
            var seriesGrp = tempEle.selectAll(`li.${SERIES_LEGEND_CLASSNAME}`).nodes(),
                series = seriesGrp[temp_i - 1],//we take from temp_i - incase of hiddenitems, all will be rendered and only for non hidden element console comes here. to access the proper element we do this.
                curr_offsetTop = !spec.data[temp_i - 1].hideInLegend ? series.offsetTop : offsetTop,
                curr_offsetLeft = !spec.data[temp_i - 1].hideInLegend ? series.offsetLeft : offsetLeft,
                expandable_ele = tempEle.selectAll(".zc-expandable").node(),
                expandable_height = isUndefined(expandable_height) ? expandable_ele && expandable_ele.offsetHeight || 0 : expandable_height,
                curr_expandable_height = expandable_ele && expandable_ele.offsetHeight || 0;

            series_width = !defined(series_width) ? series.offsetWidth : series_width;
            series_height = !defined(series_height) ? series.offsetHeight : series_height;
            prev_h = curr_offsetTop != offsetTop ? temp_h : prev_h;
            prev_w = curr_offsetLeft != offsetLeft ? temp_w : prev_w;
            temp_h = h;
            temp_w = w;
            isWidthInLimit = w < maxW;
            isHeightInLimit = h < maxH;

            offsetTop = !defined(offsetTop) ? curr_offsetTop : offsetTop;
            offsetLeft = !defined(offsetLeft) ? curr_offsetLeft : offsetLeft;

            var newRow = curr_offsetTop != offsetTop;
            if (newRow) {
                prevH_without_expandable = prev_h - expandable_height;
            }
            if (defined(prev_h) && defined(prev_w)) {
                if (h === maxH) {
                    if (prev_h + series_height <= maxH) {
                        isHeightInLimit = true;
                    }
                }
                if (w === maxW) {
                    if (prev_w + series_width <= maxW) {
                        isWidthInLimit = true;
                    }
                }

                //TODO - need to be removed
                if (w < maxW && w + series_width > maxW && prev_w < maxW && prev_w + series_width > maxW && prev_w != w && !(spec.data.length === temp_i)) {
                    //minor diff will be there say 1 sometimes - so we have this
                    isWidthInLimit = false;
                }

                //reason for below condition
                //in precalculationat one point height will got along with expandable, but for last item expandable won't be there
                //but in comparison prevH will be included, so if expandable included in that height the calculation of whether the current item has overflown or not wont be fine
                //so if incase expandable got removed, we subtract the height of expdanble from prevH.
                if (!curr_expandable_height) {
                    prev_h = prevH_without_expandable;
                    expandable_height = 0;
                }
                //prev_h + series_height > maxH --> this condition because
                //even if a legend itemis going out, the height got will be within the maxH, since maxH is applied in precalculation also
                //so inorder to know whether overflowing is happened or not, we have this condition, now since we compare with previous height, if previous height +item's height is greater than macH then it means overflowing has happend and the height is not in limit.
                if (h <= maxH && prev_h <= maxH && (newRow ? prev_h + series_height > maxH : !curr_expandable_height ? !(prev_h + series_height <= maxH) : !isWithinRange) /*&& (isHorizontal ? !(prev_h === h && curr_offsetTop === offsetTop && !isWithinRange) : true)*/) {
                    //h + series_height > maxH --> not needed,but having it to avoid any breakage
                    isHeightInLimit = false;
                }
            } else {
                prev_w = temp_w;
                prev_h = temp_h;
            }

            if (curr_offsetLeft != offsetLeft) {
                offsetLeft = curr_offsetLeft;
            }
            if (newRow) {
                //came into a new row.
                var _t = isHorizontal
                    ? isSizeLegend
                        ? isWidthInLimit
                        : (isWidthInLimit && isHeightInLimit) || isHeightInLimit
                    : isHeightInLimit;
                offsetTop = curr_offsetTop;
                rowCount = isWithinRange ? rowCount + 1 : rowCount; //only if a row is rendered completely, at the start of next row, control will come here.isWithinRange also will be the pointing to the first if current row is 2.isWithinRange will tell if previous row was fittedin or not.
                //when only one row, but not fiitedin, then we will not increment it here, at the end if rowCount is 0, we will make it to 1.
                currentRow_rendered = isWithinRange && !_t ? dummy : true;

                //at the start of 2nd row control will come in here, so we collect the itemCount in first row like below
                if (!defined(itemCountInFirstrow)) {
                    //when maxHeight is very low, by default one row has to come, but inorder to know the number of tiems in a row we do this
                    itemCountInFirstrow = temp_i - 1;
                }
            }
        } else {
            currentRow_rendered = currentRow_rendered ? currentRow_rendered : true;
            isWidthInLimit = w < maxW;
            isHeightInLimit = h < maxH;
        }
        isWithinRange = isHorizontal
            ? isSizeLegend
                ? isWidthInLimit
                : (isWidthInLimit && isHeightInLimit) || isHeightInLimit
            : isHeightInLimit;
        if (isWithinRange) {
            i = temp_i;
        }
    }
    if (!isSizeLegend) {
        if (isWithinRange) {
            // - for last row intially there will be no expandable, due to expandable complete maxHeight maynot be utilized, if all the elements rendered- no expandable, so height compied completely and will be within the range
            rowCount++;
            // i = --temp_i;
        }

        itemCountInFirstrow = !defined(itemCountInFirstrow) ? spec.data.length : itemCountInFirstrow;
        //BELOW CASE - single row can't fit in the given maxheight
        i = !i ? itemCountInFirstrow : i;
        rowCount = !rowCount ? 1 : rowCount; //also if single row, rowCount will be 0 since rowCount will be incremented only if the control goes to a new row and the previous row is fittedin(isWthinRange is true)
    } else {
        i = temp_i;
        if (isHorizontal ? w >= maxW : h >= maxH) {
            i--;
        }
    }

    titleConfig.show = actualTitleShow;
    expandableConfig._expanded = actualExpanded;

    tempEle.remove();
    //itemCount in a row will not be same for all rows - so we took another variable called rowCount.
    return {
        visibleItemCount: mathMax(i, 1),
        itemCountInFirstrow: itemCountInFirstrow,
        rowCount: rowCount
    };
}
