// @zohocharts/zc/src/legend/size-legend/list/size-list.js
//
import { Registry } from '../../../core/zc-registry';
import { LegendRegistry } from '../../legend-registry';
import { SizeLegend } from '../size-legend';
import { ListLegendParser } from '../../list-legend/list-legend-parser';
import { NULL } from '../../../core/zc-constants';
import { HORIZONTAL_ALIGN } from '../../../core/align-constants';
import { ExpandableAdjustFactor } from '../../expandable/expandable-adjust';
import { LegendTitleHtmlAdjustFactor } from '../../title/legend-title-html-adjust';
import { getLegendAlignment } from '../../legend-utils';
import { getMaxToResize } from '../../legend-bound';
import { getVisibleItemRenderCount } from './expandable-helper';

class SizeLegendList extends SizeLegend {
    create() {
        let viewScaleComponent = this.viewScaleComponent();
        this.spec.data = viewScaleComponent.ticks;
        this.spec.scales = { size: viewScaleComponent.scale };
        let _expand_info = getVisibleItemRenderCount(this);
        this.spec.visibleItemCount = _expand_info && _expand_info.visibleItemCount;

        let mark = ListLegendParser(this.spec, viewScaleComponent, this); //TODO: remove 'this'
        this._parser = mark._helper;
        return (this.encode = mark);
    }
    draw(parent) {
        parent.styles({
            display: 'block',
            position: 'absolute',
            left: NULL, //Reset left and top position. so that floating legend won't get trimmedon redraw (legend-03.json)
            top: NULL
        });
        super.draw(parent);
    }

    resize(width, height) {
        if (!width && !height) {
            return;
        }

        let viewScaleComponent = this.viewScaleComponent();
        let _expand_info = getVisibleItemRenderCount(this);
        this.spec.visibleItemCount = _expand_info && _expand_info.visibleItemCount;

        let { w, h } = getMaxToResize(width, height, this.spec.bound);
        this.updateScale(w, h);
        super.resize(w, h);
    }
}

SizeLegendList.adjustFactor = (config, w, h, globalFont) => {
    let align = getLegendAlignment(config),
        { maxWidth, maxHeight, title: titleConfig, list: listConfig } = config,
        { expandable: expandableConfig } = listConfig,
        expandableAdjust = ExpandableAdjustFactor(expandableConfig, w, h, globalFont, align, config.itemMargin),
        titleAdjust = LegendTitleHtmlAdjustFactor(titleConfig, w, h, globalFont, align),
        scrollBarSize = 22,
        adjust = [0, 0];

    if (align.layout === HORIZONTAL_ALIGN) {
        adjust[1] += scrollBarSize;
    } else {
        adjust[0] += scrollBarSize;
    }

    if (!maxHeight) {
        adjust[1] += expandableAdjust[1] + titleAdjust[1];
    }

    if (!maxWidth) {
        adjust[0] += expandableAdjust[0] + titleAdjust[0];
    }

    return adjust;
};

SizeLegendList.encode = 'z';
SizeLegendList.view = 'list';
SizeLegendList.domMode = 'html';

export { SizeLegendList };
Registry.setComponent('sizeLegendList', SizeLegendList);
LegendRegistry.add(SizeLegendList);
