// @zohocharts/zc/src/legend/color-legend/list/color-list.js
import { Registry } from '../../../core/zc-registry';
import { LegendRegistry } from '../../legend-registry';
import { ColorLegend } from '../color-legend';
// export class ColorLegendList extends ColorLegend {}

import { renderDom } from '../../../mark/dom-render';
import { NULL } from '../../../core/zc-constants';
import { ListLegend } from '../../list-legend/list-legend';
import { d3_scaleOrdinalOrg } from '../../../core/d3-constants';
import { ColorListAdapter, updateUnitTestObj } from './color-list-helper';
import { getVisibleItemRenderCount } from '../../size-legend/list/expandable-helper';

export class ColorLegendList extends ColorListAdapter {
    constructor(spec) {
        super(spec);
        spec.config = {
            ...spec.config,
            list: {
                marker: { ...spec.config.colorBox },
                label: {
                    ...spec.config
                },
                expandable: spec.config.expandable
            }
        };

        this.spec = spec;
    }
    create() {
        let depends = this.get();
        this.spec.data = depends.ticks;
        this.spec.format = depends.format;

        
        this.spec.scaleComponent = {
            getScale: () => d3_scaleOrdinalOrg().domain(this.spec.data),
            schema: {
                domain: {
                    ranges: this.spec.data
                },
                data: {
                    meta: { datatype: 'ordinal' },
                    dataType: 'ordinal'
                }
            }
        };

        this.spec.config.list.marker.fillColor = depends.color;
        
        this.spec.computedInfo = depends.computedInfo;
        this.spec.helpers = depends.helpers;

        var _expandable_info = getVisibleItemRenderCount(this);
        this.spec.visibleItemCount = _expandable_info && _expandable_info.visibleItemCount;
        this.spec.itemCountInArow = _expandable_info && _expandable_info.itemCountInArow;
        this.spec.rowCount = _expandable_info && _expandable_info.rowCount;
        let viewObj = new ListLegend(this.spec, this.viewScaleComponent);
        
        let encode = (this.encode = viewObj.create());
        this._parser = encode._helper;

        updateUnitTestObj(encode, this.spec);
        return encode;
    }
    draw(parent) {
        parent.styles({
            display: 'block',
            position: 'absolute',
            left: NULL, //Reset left and top position. so that floating legend won't get trimmedon redraw (legend-03.json)
            top: NULL
        });
        renderDom(this.encode, parent, this.spec.callbackArgs[0].viewFlow);
        this.parent = parent;
    }

    resize(width, height) {
        // this.parent.styles({
        //     height: `${height}px`,
        //     width: `${width}px`,
        //     overflow: 'hidden'
        // });
        super.resize(width, height);
    }
}

ColorLegendList.encode = 'clr';
ColorLegendList.view = 'list';
ColorLegendList.domMode = 'html';

Registry.setComponent('colorLegendList', ColorLegendList);
LegendRegistry.add(ColorLegendList);
