// @zohocharts/zc/src/legend/color-legend/list/color-list-helper.js
import { $ZCG } from '../../../core/zcglobal';
import { mathMax, mathMin, NULL, POINTER, PX, TRANSPARENT } from '../../../core/zc-constants';
import { getLegendAlignment } from '../../legend-utils';
import { helpers } from '../../legend-discrete-helpers';
import { utils, defined, getValByPriority, pick, invokeFunction, mergeJSON } from '../../../utils/zcutils';
import { $Browser } from '../../../core/browser-detect';
import { legend, legend_helpers } from '../../../core/main-zc-init';
import { getDisabledOptions } from '../../events/legend-event-helpers';
import { ColorLegend } from '../color-legend';
import { getColorByIndex } from '../../../utils/colorutils';
import { getLegendPosition } from '../../../legend/legend-utils';
import { HORIZONTAL_ALIGN, MIDDLE_ALIGN } from '../../../core/align-constants';
import { attachevents, detachevent } from '../../../events/eventActions';
import { d3_event, d3_rgb, d3_select } from '../../../core/d3-constants';
import { applyFont } from '../../../utils/zcsvgutils';
import { isFunction } from '../../../utils/function-utils';
import { Registry } from '../../../core/zc-registry';
import { collectLegendDimVals } from '../../list-legend/list-legend-parser';

let getClr = legend_helpers.getClr,
    isDisabled = legend_helpers.isDisabled;

export class ColorListAdapter extends ColorLegend {
    constructor(spec) {
        super(spec);
        this.spec = spec;
    }

    get() {
        let _this = this,
            spec = this.spec,
            { config, globalFont, chartObj } = spec,
            legData = legend_helpers.getLegendData(chartObj, spec.order),
            legData_length = legend_helpers.getLegendDataLength(chartObj, legData),
            priorityArray = [config, chartObj.systemConf.canvas, $ZCG],
            fontSize = +utils.getVal(config.fontSize || globalFont.fontSize),
            fontColor = getValByPriority(priorityArray, 'fontColor'),
            boxOptions = config.list.marker,
            selectAllBoxOptions = config.selectAllBox || {},
            colorBox = config.colorBox,
            colorBoxSize = legend_helpers.getFilterBoxSize(boxOptions, fontSize),
            isAllEnabled = legend_helpers.isAllEnabled,
            colorBoxWidth = colorBoxSize[0],
            colorBoxHeight = colorBoxSize[1],
            legendAlign = getLegendAlignment(config),
            disabledOpacity = (selectAllBoxOptions && selectAllBoxOptions.disabledOpacity) || 0.3,
            lineHeight = defined(boxOptions.size)
                ? `${mathMax(fontSize, colorBoxHeight)}px`
                : ($Browser.OSX || $Browser.LINUX) && $Browser.FIREFOX
                ? '1.2'
                : 'normal',
            titlePriorityArray = [config.title, chartObj.systemConf.canvas, $ZCG],
            { order } = spec,
            mergedColorBoxOptions = legend_helpers.getFilterBoxSize(mergeJSON(boxOptions, selectAllBoxOptions, true), fontSize),
            selectAllFilterSize =
                Math.min(mergedColorBoxOptions[0], mergedColorBoxOptions[1]) -
                (config.imagePallete || (!defined(config.shapes) || colorBox.shape === 'square') ? 0 : +colorBox.strokeWidth);//TODO apply width and height via list-parser.js code, due to this subtraction of strokeWidth we do it in here.

        // states = (spec.states = spec.states || {
        //     allSelected: true,
        //     isFirstClickAfterDeselection: false
        // }),
        chartObj.legend.state = chartObj.legend.state || {};
        let states = (chartObj.legend.state[`clr-${order}`] = chartObj.legend.state[`clr-${order}`] || {
                allSelected: true,
                isFirstClickAfterDeselection: false
            }),
            disabledOptions = getDisabledOptions(config, 'clr'),
            disabledColorBox = disabledOptions.colorBox || {},
            ImpFontProp = {
                fontColor: function(d) {
                    return getClr(isDisabled(d, states.allSelected, states.isFirstClickAfterDeselection), fontColor, disabledOptions);
                },
                textDecoration: function(d) {
                    return isDisabled(d, states.allSelected, states.isFirstClickAfterDeselection) && disabledOptions.strikeout
                        ? 'line-through'
                        : null;
                }
            },
            colorBoxOptions = spec.config.colorBox,
            colorBoxPosition = colorBoxOptions.position || 'left',
            showAs = spec.config.showAs,
            expandableConfig = config.expandable,
            legendDiscreteHelper = helpers(chartObj, null, legData, legData_length, colorBoxSize, legendAlign, lineHeight, ImpFontProp),
            computedBoxStyles = legend.setColorBoxBackground(null, chartObj, states, order, spec),
            selectAllObj = legend.selectAllBox(null, chartObj, states, order),
            selectAllStyles = selectAllObj.styles,
            selectAllSize = selectAllObj.size,
            padding_attr = `padding-${colorBoxPosition}`,
            padding_val = spec.config.textPadding,
            margin_attr = `margin-${colorBoxPosition}`,
            margin_val = selectAllSize,
            selectAllElement,
            selectAllHeight,
            titleHeight,
            keepLegendTitleAligned = config.title.keepAligned && legendAlign.layout === HORIZONTAL_ALIGN;
        selectAllBoxOptions = selectAllObj.options;
        chartObj.legend.states = chartObj.legend.state['clr-0'];
        var overflow_prop;
        if (config.textOverflow === 'wrap') {
            overflow_prop = {
                'word-wrap': 'break-word',
                'white-space': 'normal',
                'vertical-align': 'top'
            };
        } else {
            overflow_prop = {
                'text-overflow': 'ellipsis',
                'vertical-align': 'middle'
            };
        }

        states.allSelected = defined(selectAllBoxOptions.selected) ? selectAllBoxOptions.selected : states.allSelected;

        legend_helpers.fadeChart(states.allSelected, chartObj, selectAllBoxOptions.disabledOpacity);

        //legend information

        //onEnd
        var legend_onEnd = ele => {
            if (defined(showAs)) {
                ele.each(function(d, i) {
                    // #ZC1872 -- apply only if the data is disabled
                    if (isDisabled(d, states.allSelected, states.isFirstClickAfterDeselection)) {
                        var _ = applyFont(
                            d3_select(this).selectAll('.zc-legend-legendlabel, .zc-legend-legendlabel *'),
                            [
                                { fontColor: ImpFontProp.fontColor(d), textDecoration: ImpFontProp.textDecoration(d) },
                                config,
                                globalFont,
                                $ZCG
                            ],
                            true
                        );
                        if ((expandableConfig.show && defined(_this.spec.visibleItemCount)) || !expandableConfig.show) {
                            legendGrp.legend_grp.legend_nodes.disabledStyles = _;
                        }
                    }
                });
            }
        };

        //colorBox styles
        var legend_clrBoxStyles = {},
            legend_clrBoxStyles_keys = Object.keys(computedBoxStyles.styles),
            legend_clrBoxHTML;

        legend_clrBoxStyles_keys.map(key => {
            var _result = computedBoxStyles.styles[key];
            if (typeof _result === 'function') {
                _result = function(d, i) {
                    var _ = computedBoxStyles.styles[key](d, i);
                    var unitTest_colorBox = legendGrp.legend_grp.legend_nodes.colorBox;
                    var _styles = unitTest_colorBox.seriesSpecificStyles;
                    _styles[i] = _styles[i] ? _styles[i] : {};

                    if (key === 'background-size') {
                        _styles[i].backgroundSize = _;
                    } else {
                        _styles[i][key] = _;
                    }
                    return _;
                };
            }
            legend_clrBoxStyles[key] = _result;
        });

        legend_clrBoxHTML = function(d, i) {
            var _ = computedBoxStyles.html(d, i);

            if ((expandableConfig.show && defined(_this.spec.visibleItemCount)) || !expandableConfig.show) {
                var _obj = legendGrp.legend_grp.legend_nodes.colorBox.seriesSpecificStyles;
                _obj[i] = _obj[i] ? _obj[i] : {};
                _obj[i].html = _;
            }
            return _;
        };

        //label styles
        var legend_labelStyles = {
            ...applyFont(null, [ImpFontProp, config, globalFont, $ZCG], true)
        };
        if(!$Browser.PHANTOMJS){
            legend_labelStyles["vertical-align"] =  "inherit";
        }
        legend_labelStyles[padding_attr] = padding_val;
        legend_labelStyles[margin_attr] = colorBoxWidth;
        if (defined(showAs)) {
            var _s = {
                // #ZC2002
                display: pick(spec.config.showAsDisplay, 'inline-block'),
                overflow: 'hidden'
            };
            legend_labelStyles = { ...legend_labelStyles, ..._s };
        }

        //Title information
        //onEnd
        var title_onEnd = ele => {
                ele.style('line-height', function() {
                    var selectAllFilterHeight = selectAllBoxOptions.show ? selectAllHeight : 0,
                        headerLineHeight = Math.max(titleHeight, selectAllFilterHeight);
                    if ((expandableConfig.show && defined(_this.spec.visibleItemCount)) || !expandableConfig.show) {
                        legendGrp.title_grp.title_nodes.lineHeight = headerLineHeight;
                    }
                    return `${headerLineHeight}px`;
                });
                function selectBoxClickFunction(e) {
                    if (!(states.allSelected && !isAllEnabled(legData, states.allSelected))) {
                        states.allSelected = !states.allSelected;
                    }
                    states.isFirstClickAfterDeselection = false;
                    selectAllElement.html(
                        states.allSelected ? pick(selectAllObj.options.text, colorBox.text) || '&#10003;' : disabledColorBox.text || ''
                    );
                    var userFunction = selectAllObj.options.events[chartObj.zcEventType || 'click'];
                    invokeFunction(userFunction, d3_event(e), states.allSelected, chartObj);
                }
                var mouseHandlers = {
                        click: selectBoxClickFunction
                    },
                    touchHandlers = {
                        tap: selectBoxClickFunction
                    };

                if (selectAllBoxOptions.show) {
                    attachevents(ele, mouseHandlers, touchHandlers);
                } else {
                    detachevent(ele, null, mouseHandlers);
                    detachevent(ele, null, touchHandlers);
                }
            },
            title_selectAll_onEnd = ele => {
                selectAllElement = ele;
                selectAllHeight = ele.node() ? ele.node().offsetHeight : selectAllHeight;

                ele.styles({
                    width: `${selectAllFilterSize}px`,
                    height: `${selectAllFilterSize}px`
                });
            },
            title_label_onEnd = ele => {
                const node = ele.node();
                titleHeight = node ? node.offsetHeight || node.getBoundingClientRect().height : titleHeight;    // #ZC4875 - Safari browser offsetHeight issue fix
                selectAllElement.style('margin', function() {
                    var top = `${mathMax(0, (titleHeight - selectAllHeight) / 2)}px`;
                    if ((expandableConfig.show && defined(_this.spec.visibleItemCount)) || !expandableConfig.show) {
                        legendGrp.title_grp.title_nodes.selectAll.margin = `${top} 0px 0px 0px`;
                    }
                    return `${top} 0px 0px 0px`;
                });
                //for unit testing, since in old config these are maintained, we need to store it here also.
                // if(!selectAllElement.node()){
                //     var top = `${mathMax(0, (titleHeight - selectAllHeight) / 2)}px`;
                //     if ((expandableConfig.show && defined(_this.spec.visibleItemCount)) || !expandableConfig.show) {
                //         legendGrp.title_grp.title_nodes.selectAll.margin = `${top} 0px 0px 0px`;
                //     }
                // }
            },
            titleGrp_onEnd = ele => {
                var { bound, config } = spec,
                    titleConfig = config.title,
                    keepTitleAligned =
                        titleConfig && titleConfig.show && titleConfig.keepAligned && legendAlign.layout === HORIZONTAL_ALIGN,
                    wrap = d3_select(ele.node().parentNode),
                    showAs = config.showAs,
                    isAutoItemWidth = config.itemWidth === 'auto',
                    { maxW } = collectLegendDimVals(
                        wrap,
                        legendAlign,
                        keepTitleAligned,
                        bound,
                        config,
                        legendDiscreteHelper,
                        legData,
                        spec,
                        globalFont
                    ),
                    legComp = chartObj.legend.component,
                    encode = this.discrete_encode;
                
                encode.legendLabelMaxWidth = legComp.legendLabelMaxWidth = null;
                encode.itemMaxWidth = legComp.itemMaxWidth = null;
                if (showAs && isAutoItemWidth) {
                    let maxItemWidth = legendDiscreteHelper.getMaxItemWidth(legData),
                        constrainedMaxWidth = mathMin(maxItemWidth, maxW),
                        fontSize = +utils.getVal(config.fontSize || globalFont.fontSize),
                        colorBoxSize = legend_helpers.getFilterBoxSize(config.list.marker, fontSize);

                    encode.itemMaxWidth = legComp.itemMaxWidth = spec.itemMaxWidth = constrainedMaxWidth;
                    // #ZC1904 subtract the itemMargin
                    encode.legendLabelMaxWidth = legComp.legendLabelMaxWidth = spec.legendLabelMaxWidth = constrainedMaxWidth - colorBoxSize[0] - config.itemMargin * 2;
                    wrap.selectAll('.zc-legend-legendlabel').style('width', spec.legendLabelMaxWidth + PX);
                }
            };

        //selectAll styles
        var title_selectAll_color = function(d, i) {
                var _c = selectAllStyles.color(d, i);
                if ((expandableConfig.show && defined(_this.spec.visibleItemCount)) || !expandableConfig.show) {
                    legendGrp.title_grp.title_nodes.selectAll.color[i] = _c;
                }

                return _c;
            },
            title_selectAll_html = function(d, i) {
                var _html = selectAllObj.html(d, i);
                if ((expandableConfig.show && defined(_this.spec.visibleItemCount)) || !expandableConfig.show) {
                    legendGrp.title_grp.title_nodes.selectAll.html = _html;
                }
                return _html;
            };

        //label styles
        var title_labelStyles = {
            ...applyFont(null, [config.title, globalFont, $ZCG], true)
        };
        title_labelStyles[padding_attr] = selectAllObj.options.show ? padding_val + 'px' : null;
        title_labelStyles[margin_attr] = selectAllObj.options.show ? margin_val + 'px' : null;

        function alignLegend(container, floating = false, maxWidth, maxHeight, chartarea) {
            let x = 0,
                y = 0;
            if (floating) {
                let legend = systemconf.legend;
                var dimension = chartObj.legendarea.element.node().getBoundingClientRect();
                x = mathMin(chartObj.canvasarea.width - dimension.width, legend.x);
                y = mathMin(chartObj.canvasarea.height - mathMin(dimension.height, maxHeight), legend.y);
            } else {
                let position = getLegendPosition(chartarea, chartObj);
                x = position[0];
                y = position[1];
            }
            container.styles({
                left: `${x}px`,
                top: `${y}px`
            });
        }

        var legendGrp = constructUnitTestObject();
        legend_labelStyles[padding_attr] = legend_labelStyles[padding_attr]
            ? legend_labelStyles[padding_attr] + 'px'
            : legend_labelStyles[padding_attr];
        legend_labelStyles[margin_attr] = legend_labelStyles[margin_attr]
            ? legend_labelStyles[margin_attr] + 'px'
            : legend_labelStyles[margin_attr];

        var computedParserInfo = getComputedParserInfo();

        function initLegendComp(ele){
            _this.discrete_encode = _this.spec.chartObj.legend.component = legend.discrete.get(_this.spec);
            legend.selectAllBox(null, chartObj, chartObj.legend.state[spec.id], order);
            legend.events(chartObj, ele, chartObj.legend.state[spec.id], disabledOpacity, order);
        }
        function getComputedParserInfo() {
            return {
                initLegendComp: initLegendComp,
                expandable: {
                    alignLegend: alignLegend
                },
                title: {
                    marker: {
                        styles: {
                            ...selectAllStyles,
                            color: title_selectAll_color,
                            'background-repeat': null
                        },
                        config: {
                            background: selectAllStyles.background,
                            border: selectAllStyles.border,
                            html: title_selectAll_html //,
                            // tooltip: selectAllBoxOptions.tooltip.enabled ? true : false//inorder to show tooltip
                        },
                        onEnd: title_selectAll_onEnd
                    },
                    label: {
                        styles: {
                            ...title_labelStyles
                        },
                        config: {
                            strokeColor: getValByPriority(titlePriorityArray, 'strokeColor')
                        },
                        onEnd: title_label_onEnd
                    },
                    li: {
                        styles: {
                            ...applyFont(NULL, [], true),
                            margin: null,
                            'vertical-align': MIDDLE_ALIGN,
                            'text-overflow': 'ellipsis',
                            position: null,
                            height: null,
                            'word-wrap': null,
                            display: null,
                            cursor: selectAllBoxOptions.show ? POINTER : null,
                            'font-size': null
                        },
                        onEnd: title_onEnd
                    },
                    ul: {
                        styles: {
                            'white-space': 'normal',
                            display: keepLegendTitleAligned ? 'inline-block' : null,
                            overflow: 'hidden',
                            '-webkit-overflow-scrolling': null,
                            'font-size': null,
                            filter: null
                        },
                        onEnd: titleGrp_onEnd
                    }
                },
                legend: {
                    marker: {
                        styles: {
                            'line-height': colorBoxHeight - colorBoxOptions.strokeWidth + PX,
                            // 'font-family': colorBoxOptions.fontFamily,
                            // 'font-size': (colorBoxOptions.fontSize || fontSize) + PX,
                            'border-color': TRANSPARENT,
                            'background-repeat': 'repeat',
                            color: legend_clrBoxStyles.color,
                            filter: legend_clrBoxStyles.filter,
                            opacity: legend_clrBoxStyles.opacity,
                            display: 'inline-block',
                            float: colorBoxPosition,
                            position: 'absolute',
                            'box-sizing': 'border-box',
                            'text-align': 'center',
                            'font-family': colorBoxOptions.fontFamily,
                            'font-size': (colorBoxOptions.fontSize || fontSize) + PX,
                            margin: null
                        },
                        config: {
                            //giving below in config, since we have used _div and default div configd will be applied via renderConfig, so if we give configs in here, then these will be taken over the default ones.
                            background: legend_clrBoxStyles.background,
                            backgroundSize: legend_clrBoxStyles['background-size'],
                            border: legend_clrBoxStyles.border,
                            borderRadius: legend_clrBoxStyles['border-radius'],
                            html: legend_clrBoxHTML
                            //width and height given in config, since by default a width and height is given in config, and inorder to override it we give this in config
                            // width: colorBoxWidth,
                            // height: colorBoxHeight
                        }
                    },
                    label: {
                        styles: {
                            margin: null,
                            ...legend_labelStyles
                        },
                        config: {
                            order: null,
                            tooltip: config.tooltip ? true : false //inorder to show tooltip
                        }
                    },
                    li: {
                        styles: {
                            ...applyFont(NULL, [], true), //since in list-parser - while collecting li spec, applyFOnt is called, since if removed size legend will get affected added dummy values here and for title li also added.
                            color: function(d, i) {
                                var _color = getClr(
                                    isDisabled(d, states.allSelected, states.isFirstClickAfterDeselection),
                                    fontColor,
                                    disabledOptions
                                );
                                if ((expandableConfig.show && defined(_this.spec.visibleItemCount)) || !expandableConfig.show) {
                                    var _obj = legendGrp.legend_grp.legend_nodes.seriesSpecificStyles;
                                    _obj[i] = _obj[i] ? _obj[i] : {};
                                    _obj[i].color = _color;
                                }
                                return _color;
                            },
                            'line-height': lineHeight,
                            'font-size': `${fontSize}px`,
                            'font-family': getValByPriority(priorityArray, 'fontFamily'),
                            'font-weight': getValByPriority(priorityArray, 'fontWeight'),
                            'box-sizing': 'border-box'
                        },
                        onEnd: legend_onEnd
                    },
                    ul: {
                        styles: {
                            'white-space': 'normal',
                            'overflow-x': 'hidden'
                        }
                    }
                }
            };
        }

        function constructUnitTestObject() {
            chartObj.legend.unitTestObject = {
                'zc-legend': {
                    clr: {}
                }
            };
            var _ = (chartObj.legend.unitTestObject['zc-legend'].clr[order] = {
                fontSize: fontSize,
                fontColor: fontColor,
                background_rgb: d3_rgb(config.background.color.toString()),
                title_grp: {
                    title_nodes: {
                        selectAll: {
                            ...selectAllStyles,
                            color: {},
                            cursor: selectAllBoxOptions.show ? POINTER : null
                            //cursor was applied to li element not selectAll, but we put it here since in the reference we put it under li, so inorder to maintain it we do this
                        },
                        title: {
                            ...title_labelStyles,
                            html_or_text: config.title.text
                        }
                    },
                    display: keepLegendTitleAligned ? 'inline-block' : null
                },
                legend_grp: {
                    legend_nodes: {
                        //cursor will get applied only on hover over the legend, check show-cursor, hide-cursor files.
                        cursor: config.cursor || POINTER,
                        ...overflow_prop,
                        'vertical-align': 'top',
                        lineHeight: lineHeight,
                        'font-family': getValByPriority(priorityArray, 'fontFamily'),
                        'font-weight': getValByPriority(priorityArray, 'fontWeight'),
                        seriesSpecificStyles: {},
                        label: {
                            disabledStyles: {},
                            ...legend_labelStyles,
                            color: fontColor,
                            'text-decoration': null,
                            title: {}
                        },
                        colorBox: {
                            position: colorBoxPosition,
                            size: colorBoxSize,
                            lineHeight: colorBoxHeight - colorBoxOptions.strokeWidth,
                            seriesSpecificStyles: {},
                            borderRadius: legend_clrBoxStyles['border-radius']
                        }
                    }
                }
            });
            if (config.title.show) {
                _.title_grp.title_nodes.selectAll.html = selectAllObj.html();
            }

            return _;
        }
        return {
            ticks: legData,
            format: legendDiscreteHelper.getLegendItemText,
            color: (d, i) => getColorByIndex(chartObj, i, null, order),
            computedInfo: computedParserInfo,
            helpers: legendDiscreteHelper
        };
    }
}

export function updateUnitTestObj(parser, spec) {
    var chart = spec.chartObj,
        unitTestObj = chart.legend.unitTestObject['zc-legend'].clr[spec.order],
        markSelection = Registry.getComponent('markSelection'),
        markSelector = new markSelection(parser),
        unitTest_selectAll = unitTestObj.title_grp.title_nodes.selectAll,
        parser_selectAll = markSelector.selectAll('selector', '.zc-header-legendfilterbox').selection[0],
        selectAll_width = parser_selectAll.config.width,
        selectAll_height = parser_selectAll.config.height,
        unitTest_title = unitTestObj.title_grp.title_nodes.title,
        parser_title = markSelector.selectAll('selector', '.zc-header-legendlabel').selection[0],
        title_html = parser_title.config.html,
        unitTest_titleLI = unitTestObj.title_grp.title_nodes,
        parser_titleLI = markSelector.selectAll('selector', '.zc-header-series').selection[0],
        titleLI_padding = parser_titleLI.properties.styles.padding,
        unitTest_titleUL = unitTestObj.title_grp,
        parser_titleUL = markSelector.selectAll('selector', '.zc-header-legseries').selection[0],
        titleUL_verticalAlign = parser_titleUL.properties.styles['vertical-align'],
        unitTest_legendLabel = unitTestObj.legend_grp.legend_nodes.label,
        parser_legendLabel = markSelector.selectAll('selector', '.zc-legend-legendlabel').selection[0],
        legendLabel_html = parser_legendLabel.config.html,
        unitTest_colorBox = unitTestObj.legend_grp.legend_nodes.colorBox,
        parser_colorBox = markSelector.selectAll('selector', '.zc-legend-legendfilterbox').selection[0],
        colorBox_html = parser_colorBox.config.html,
        unitTest_legendLI = unitTestObj.legend_grp.legend_nodes,
        parser_legendLI = markSelector.selectAll('selector', '.zc-legend-series').selection[0],
        legendLI_width = parser_legendLI.properties.styles.width,
        legendLI_height = parser_legendLI.properties.styles.height,
        legendLI_display = parser_legendLI.properties.styles.display,
        unitTest_legendUI = unitTestObj.legend_grp,
        parser_legendUI = markSelector.selectAll('selector', '.zc-legend-legseries').selection[0],
        legendUL_verticalAlign = parser_legendUI.properties.styles['vertical-align'];

    //selectAll
    parser_selectAll.config.width = function(d, i) {
        var _result = selectAll_width.call(this, d, i);
        unitTest_selectAll.width = _result + 'px';
        return _result;
    };
    parser_selectAll.config.height = function(d, i) {
        var _result = selectAll_height.call(this, d, i);
        unitTest_selectAll.height = _result + 'px';
        return _result;
    };

    //title label
    parser_title.config.html = function(d, i) {
        var _r = title_html.call(this, d, i),
            titleContent = _r;
        if (isFunction(_r)) {
            titleContent = invokeFunction(_r, chart);
        }
        unitTest_title.html_or_text = titleContent;
        return titleContent;
    };

    //title li
    unitTest_titleLI.padding = titleLI_padding;

    //title ul
    unitTest_titleUL['vertical-align'] = titleUL_verticalAlign;

    //legend label
    unitTest_legendLabel.seriesSpecificHTML = {};
    parser_legendLabel.config.html = function(d, i) {
        var _result = legendLabel_html.call(this, d, i);
        unitTest_legendLabel.seriesSpecificHTML[i] = _result;
        return _result;
    };

    //colorbox
    unitTest_colorBox.seriesSpecificStyles = unitTest_colorBox.seriesSpecificStyles ? unitTest_colorBox.seriesSpecificStyles : {};
    parser_colorBox.config.html = function(d, i) {
        var _result = colorBox_html.call(this, d, i);
        unitTest_colorBox.seriesSpecificStyles[i].html = _result;
        return _result;
    };

    //legendLi
    unitTest_legendLI.width = legendLI_width;
    unitTest_legendLI.height = legendLI_height;

    unitTest_legendLI.seriesSpecificStyles = unitTest_legendLI.seriesSpecificStyles ? unitTest_legendLI.seriesSpecificStyles : {};
    parser_legendLI.properties.styles.display = function(d, i) {
        var _result = legendLI_display(d, i);
        unitTest_legendLI.seriesSpecificStyles[i] = unitTest_legendLI.seriesSpecificStyles[i]
            ? unitTest_legendLI.seriesSpecificStyles[i]
            : {};
        unitTest_legendLI.seriesSpecificStyles[i].display = _result;
        return _result;
    };

    //legendUL
    unitTest_legendUI['vertical-align'] = legendUL_verticalAlign;

    //expandable
    var parser_expandable = markSelector.selectAll('selector', '.zc-expandable').selection[0];
    var expandable_html = parser_expandable.config.html;
    parser_expandable.config.html = function(d, i) {
        var _result = expandable_html.call(this, d, i);
        unitTestObj.expandable.html = _result;
        return _result;
    };
}
