// @zohocharts/zc/src/core/data-processor/transforms/sankey/sankey-layout.js
import { defined, getDistinct } from '../../../../utils/zcutils';
import { d3_extent, d3_map, d3_set } from '../../../d3-constants';
import { mathAbs, mathMax, mathMin } from '../../../zc-constants';
import { Registry } from '../../../zc-registry';

export class SankeyLayout {
    /**
     *
     * @param source
     * @param target
     * @param weight
     * @return {boolean}
     */
    static isValidPoint(source, target, target1, weight) {
        return defined(weight) && weight > 0 && defined(source) && defined(target1) && source !== target;
    }

    /**
     *
     * @param points
     * @param {Accessors} accessors
     * @param excludeItems
     */
    constructor(points, accessors, excludeItems = d3_map()) {
        let nodes = [],
            links = [],
            linkSet = d3_set(),
            nodeSet = d3_set(),
            allNodes = [],
            allLinks = [];

        const { source: sourceAccessor,
            target: targetAccessor,
            weight: weightAccessor,
            exit: exitAccessor,
            color: colorAccessor = (point) => point.color,
            filter: filterAccessor = () => false
         } = accessors;

        points.forEach(point => {
            let source = sourceAccessor(point),
                target = targetAccessor(point),
                weight = weightAccessor(point),
                target1 = exitAccessor(point);

            if (SankeyLayout.isValidPoint(source, target, target1, weight)) {
                let sourceLevel = '', targetLevel = '', sourceFormat = 'x', targetFormat = 'y',
                    sourceAxisOrder = 0, targetAxisOrder = 0, linkDisabled = point.disabled;
                if (point.level) {
                    //TODO: Make it completely customizable based on data
                    sourceLevel = point.level.slice(0, point.level.lastIndexOf('_'));
                    targetLevel = point.level;
                    sourceFormat = 'y';
                    targetFormat = point.height > 0 ? 'y' : 'x';
                    sourceAxisOrder = point.depth - 2;
                    targetAxisOrder = point.height > 0 ? point.depth - 1 : 0;
                }

                let sourceNode = {
                    name: `${source}`, nameOriginal: sourceAccessor(point, true), point, _type: 'node', 
                    _format: sourceFormat, level: sourceLevel, _axisOrder: sourceAxisOrder, 
                    color: colorAccessor(point, 'source')
                    
                },
                    targetNode = {
                        name: `${target}`, nameOriginal: targetAccessor(point, true), point, _type: 'node', 
                        _format: targetFormat, level: targetLevel, _axisOrder: targetAxisOrder, 
                        color: colorAccessor(point, 'target')
                    },
                    link = {
                        _type: 'link',
                        source: `${source}`,
                        target: `${target}`,
                        value: weight,
                        exit: !defined(target),
                        level: targetLevel,
                        color: colorAccessor(point, 'link'),
                        point
                    };

                if (!excludeItems.has(source) && !excludeItems.has(target) && !filterAccessor(point, 'link')) {
                    links.push(link);
                    linkSet.add(source);
                    linkSet.add(target);
                    nodeSet.add(source);
                    nodes.push(sourceNode);
                    if (defined(target)) {
                        nodeSet.add(target);
                        nodes.push(targetNode);
                    } else {
                        link.exit = true;
                    }
                } else if (!excludeItems.has(source) && !filterAccessor(point, 'source')) {
                    nodeSet.add(source);
                    nodes.push(sourceNode);
                } else if (target && !excludeItems.has(target) && !filterAccessor(point, 'target')) {
                    nodes.push(targetNode);
                    nodeSet.add(target);
                }

                // add to all node sets
                allLinks.push({ ...link });
                allNodes.push({ ...sourceNode });
                if (defined(target)) {
                    allNodes.push({ ...targetNode });
                }
            }
        });

        // get and set the extent
        let extent = d3_extent(links, link => link.value);

        // now find the unique nodes
        nodes = getUnique(nodes, d => d.name);
        allNodes = getUnique(allNodes, d => d.name);

        nodes.forEach(node => {
            // link less node
            if (!linkSet.has(node.name)) {
                // fixedValue a value is required when there is no link
                node.fixedValue = extent[0] || 0.0001;
            }
            node._type = 'node';
        });

        this.extent = extent;
        this.nodeLength = nodes.length;
        this.linkLength = links.length;

        this.graph = { nodes, links };
        this.completeGraph = { nodes: allNodes, links: allLinks };
    }

    /**
     *
     * @param node
     */
    getConnectedLinks(node) {
        return [...node.sourceLinks, ...node.targetLinks];
    }

    /**
     *
     * @param link
     * @return {*[]}
     */
    getConnectedNodes(link) {
        return [link.source, link.target];
    }

    /**
     *
     * @param point
     * @param flatten
     * @return {(*)[]|{nodes: [*, *], links: [*]}}
     */
    getAllConnectedPoints(point, flatten) {
        let layout = this,
            points;

        if (point._type === 'link') {
            if (flatten) {
                points = [...layout.getConnectedNodes(point), point];
            } else {
                points = { nodes: layout.getConnectedNodes(point), links: [point] };
            }
        } else if (point._type === 'node') {
            let sourceLinkNodes = point.sourceLinks.map(link => link.target),
                targetLinkNodes = point.targetLinks.map(link => link.source);

            let links = layout.getConnectedLinks(point),
                nodes = getDistinct([...targetLinkNodes, ...sourceLinkNodes, point]);

            if (flatten) {
                points = [...links, ...nodes];
            } else {
                points = { nodes, links };
            }
        }
        return points;
    }
}

/**
 *
 * @param array
 * @param accessor
 * @return {[]}
 */
function getUnique(array, accessor) {
    var flags = {},
        output = [],
        l = array.length,
        i;
    for (i = 0; i < l; i++) {
        let value = accessor(array[i]);
        if (flags[value]) {
            continue;
        }
        flags[value] = true;
        output.push(array[i]);
    }
    return output;
}

Registry.setComponent('sankeyLayout', SankeyLayout);

/**
 * Accessors for sankey layout
 * @typedef {object} Accessors - It contains source, target, weight and color accessors for sankey layout
 * @property {Function} Accessors.source - source accessor for sankey layout
 * @property {Function} Accessors.target - target accessor for sankey layout
 * @property {Function} Accessors.exit - exit accessor for sankey layout
 * @property {Function} Accessors.weight - weight accessor for sankey layout
 * @property {Function} Accessors.color - color accessor for sankey layout
 */
