// @zohocharts/zc/src/plot/morph/morphTools.js
import { GROUP_SELECTOR } from './seriesGroupMapping';
import { d3_map } from '../../core/d3-constants';
import { getRawColor } from '../../utils/colorutils';
import { pathModifier } from './pathModifier';
import { getSvgTransform } from '../../utils/svg-transforms';

export function collectExistingSeriesInfo(chartObj, chart_types) {
    var oldseries = d3_map();
    chart_types.map((d, i) => {
        var seriesInfo = {};
        var isMultiColoring = chartObj.renderer[i].commonRendererProp().params.commonPlotOptions.multiColoring;
        seriesInfo.isAxisCategory = chartObj.isAxisCategory;
        seriesInfo.seriesIndex = i;
        seriesInfo.chartType = d.name;
        seriesInfo.pathData = collectExistingPath(chartObj, seriesInfo.isAxisCategory, d.name, i);
        seriesInfo.color = collectExistingColor(chartObj, seriesInfo.isAxisCategory, isMultiColoring, seriesInfo.pathData, i);
        seriesInfo.plotarea = chartObj.plotarea;
        seriesInfo.transformationInfo = collectTransformationInfo(chartObj, seriesInfo.isAxisCategory, d.name, i);
        oldseries.set(i, seriesInfo);
    });
    chartObj.oldSeriesInfo = oldseries;
}

var transformGroupSelector = d3_map();
transformGroupSelector.set('bar', '.zc_bar');
transformGroupSelector.set('pie', '.zc_pie');
transformGroupSelector.set('funnel', '.zc_funnel');
transformGroupSelector.set('pyramid', '.pyramidsection');
transformGroupSelector.set('scatter', '.scatter');

function collectTransformationInfo(chartObj, isAxisCategory, chartType, i) {
    var seriesholder = chartObj.svg.selectAll('.seriesholder'),
        selectorIndex = isAxisCategory ? i : '',
        transformedGroups = transformGroupSelector.get(chartType),
        transformationInfo = [];

    seriesholder
        .selectAll(`.${chartType}${GROUP_SELECTOR.get(chartType)}${selectorIndex}`)
        .selectAll(transformedGroups)
        .each(function() {
            var transform = getSvgTransform(this);
            transformationInfo.push({x: transform.translate[0], y: transform.translate[1]});
        });
    return transformationInfo;
}

function collectExistingPath(chartObj, isAxisCategory, chartType, i) {
    var seriesholder = chartObj.svg.selectAll('.seriesholder');
    var paths_old = [];
    var selectorIndex = isAxisCategory ? i : '';
    seriesholder
        .selectAll(`.${chartType}${GROUP_SELECTOR.get(chartType)}${selectorIndex}`)
        .selectAll('path')
        .each(function() {
            paths_old.push(this);
        });
    return pathModifier.get(chartType)(paths_old);
}

function collectExistingColor(chartObj, isAxisCategory, isMultiColoring, oldPaths, i) {
    var colors_old,
        selectorIndex = isAxisCategory ? i : '';
    if ((isAxisCategory && isMultiColoring) || !isAxisCategory) {
        colors_old = oldPaths.map((d, i) => {
            return getRawColor(chartObj, i, d, selectorIndex);
        });
    } else {
        colors_old = oldPaths.map(d => {
            return getRawColor(chartObj, 0, d, selectorIndex);
        });
    }
    return colors_old;
}

export function getOffsetValue(chartObj, index) {
    var oldSeriesData = chartObj.oldSeriesInfo.get(index);
    if (!chartObj.isAxisCategory) {
        if (!chartObj.isAxisCategory && !oldSeriesData.isAxisCategory) {
            return {
                x: oldSeriesData.plotarea.centerX - chartObj.plotarea.centerX,
                y: oldSeriesData.plotarea.centerY - chartObj.plotarea.centerY
            };
        } else {
            return {
                x: -chartObj.plotarea.centerX + (oldSeriesData.plotarea.left - chartObj.plotarea.left),
                y: -chartObj.plotarea.centerY - (chartObj.plotarea.top - oldSeriesData.plotarea.top)
            };
        }
    } else {
        if (chartObj.isAxisCategory && oldSeriesData.isAxisCategory) {
            return {
                x: oldSeriesData.plotarea.left - chartObj.plotarea.left,
                y: oldSeriesData.plotarea.top - chartObj.plotarea.top
            };
        } else {
            return {
                x: -(chartObj.plotarea.left - oldSeriesData.plotarea.left) + oldSeriesData.plotarea.centerX,
                y: -(chartObj.plotarea.top - oldSeriesData.plotarea.top) + oldSeriesData.plotarea.centerY
            };
        }
    }
}
