// @zohocharts/zc/src/scale-parser/axis/numeric-axis-parser.js
//$Id$
//numeric-axis-parser.js
import { defined, getPropVal, isObjectPropDefined, isUndefined, pick } from '../../utils/zcutils';
import { getAxisObject } from '../../axis/axisutils';
import { getThresholdValues } from '../../threshold/threshold-utils';
import { DataProcessor } from '../../core/data-processor';
import { d3_merge, d3_set } from '../../core/d3-constants';
import { NUMERIC_DATATYPE, TIME_DATATYPE } from '../../core/datatype-constants';
import { getAxisData } from '../../axis/axisutils';
import { _hasDefaultBase } from '../../core/data-processor/minMaxYValues';
import { LOG } from '../../core/scale-type-constant';
import { chartcategory, stack } from '../../core/main-zc-init';
import { getCorrectionParser } from '../../scale-parser/axis/correction/correction-parser';
import { modifyCorrectionOperator } from '../../scale/evaluators/correction/correction-adapter'; // TODO: Remove at end of Code modularization
import { _modularized } from '../../core/zcconf';
import { Registry } from '../../core/zc-registry';

export function getAvailableOperatorsForXAxis(scaleType, needOnlyExtremes, needOnlyCorrection) {
    var defaultOperators = d3_set([
            'brushExtremes',
            'extremes',
            'threshold',
            'bubbleBound',
            'markerBound',
            'fontBound',
            'correctionFactor'
        ]),
        correctionOperators = ['correctionFactor', 'extraCorrectionFactor'];

    if (!needOnlyExtremes) {
        defaultOperators.add('extraCorrectionFactor');
    }
    if (needOnlyCorrection) {
        defaultOperators = d3_set(correctionOperators);
    }
    if (scaleType == LOG) {
        defaultOperators.add('logConstant');
    }
    return defaultOperators;
}
export function getAvailableOperatorsForYAxis(scaleType, needOnlyExtremes, needOnlyCorrection) {
    var defaultOperators = d3_set([
        'baseLine',
        'NodataExtremes',
        'brushExtremes',
        'waterfallProcessor',
        'stackExtremes',
        'seriesExtremes',
        'levelMarker',
        'threshold',
        'min',
        'max',
        'hasDefaultBase',
        'updateBaseValue',
        'hasDefaultBaseValue',
        'bubbleBound',
        'markerBound',
        'correctionFactor',
        'fontBound',
        'waterfallBound'
    ]);
    if (needOnlyCorrection) {
        defaultOperators = d3_set(['correctionFactor', 'extraCorrectionFactor']);
    }
    if (scaleType == LOG) {
        defaultOperators.add('logConstant');
    }

    return defaultOperators;
}
export function getAvailableOperatorsForTimeXAxis(scaleType, needOnlyExtremes, needOnlyCorrection) {
    var extremesOperators = ['brushExtremes', 'seriesExtremes', 'threshold', 'min', 'max', 'correctionFactor', 'timeFormat'],
        correctionOperators = ['extraCorrectionFactor', 'bubbleBound', 'fontBound', 'markerBound', 'subfunction', 'basevalue'],
        defaultOperators = needOnlyExtremes
            ? extremesOperators
            : needOnlyCorrection
            ? correctionOperators
            : d3_merge([extremesOperators, correctionOperators]);
    if (scaleType == LOG) {
        defaultOperators.push('logConstant');
    }
    return d3_set(defaultOperators);
}
export function getAvailableOperatorsForTimeYAxis(scaleType, needOnlyExtremes, needOnlyCorrection) {
    var extremesOperators = [
            'baseLine',
            'NodataExtremes',
            'brushExtremes',
            'waterfallProcessor',
            'stackExtremes',
            'seriesExtremes',
            'levelMarker',
            'threshold',
            'min',
            'max',
            'updateBaseValue',
            'hasDefaultBaseValue',
            'correctionFactor',
            'waterfallBound',
            'timeFormat'
        ],
        correctionOperators = [
            'extraCorrectionFactor',
            'basevalue',
            'subfunction',
            'bubbleBound',
            'correctionFactor',
            'fontBound',
            'markerBound'
        ],
        defaultOperators = needOnlyExtremes
            ? extremesOperators
            : needOnlyCorrection
            ? correctionOperators
            : d3_merge([extremesOperators, correctionOperators]);
    if (scaleType == LOG) {
        defaultOperators.push('logConstant');
    }
    return d3_set(defaultOperators);
}
export function numericAxisParser(
    encodeType,
    index = 0,
    chart,
    correctionFactor,
    excludeDatalabels,
    applyZoom,
    needOnlyExtremes,
    minmaxCorrection,
    customExtent
) {
    index = index || 0;
    let domainOperators = {};
    //Extremes
    let extremesOperators = [],
        correctionOperators = [];
    var processor = chart.dataObject,
        axisConfig = getAxisData(chart, encodeType, index, null, chart.userdata),
        hasNegativeValue = processor.hasNegativeValue,
        baseline = processor.baseline,
        savedRange = encodeType == 'x' ? processor.xRange : processor.yRange[index],
        dataSet = chart.dataset,
        isPercentChart = processor.isPercentChart,
        stacked = processor.isstacked() || processor.isPercentChart,
        dataType = dataSet.getDataType(encodeType, index),
        columnInfo = dataSet.getColumn(encodeType, index),
        metaConf = columnInfo[columnInfo.datatype],
        subfunction = metaConf && metaConf.subfunction,
        encodeData = processor.getEncode(encodeType, index),
        zoomed = isObjectPropDefined(chart.systemConf, `chart.brush.${encodeType}[${index}]`) && applyZoom,
        axisObj = getAxisObject(chart, encodeType, index),
        noCorrection = dataType == TIME_DATATYPE && needOnlyExtremes,
        needOnlyCorrection = defined(customExtent);

    correctionFactor =
        columnInfo.scaleType === LOG
            ? 0
            : encodeType == 'x'
            ? pick(correctionFactor, 0.05)
            : pick(correctionFactor, chart.datalabels.hasFixedPosition ? 0.00000001 : 0.05);
    minmaxCorrection = pick(minmaxCorrection, correctionFactor);

    var defaultOperator = getDefaultOperator(encodeType, dataType, needOnlyExtremes, needOnlyCorrection, columnInfo.scaleType);

    // Pop unnecessary correction
    let toBePopped = getOperatorsForPop(encodeType, index, chart, excludeDatalabels, minmaxCorrection);
    toBePopped.forEach(_ => defaultOperator.remove(_));

    //NodataExtremes operator
    if (defaultOperator.has('NodataExtremes') && !encodeData) {
        if (needOnlyExtremes) {
            //TODO PARSER check why this special case for needonlyExtremes
            extremesOperators.push(['NoData', 'NoData']);
        } else {
            extremesOperators.push([0, 0]);
        }

        domainOperators.extremes = extremesOperators;
        return domainOperators;
    }

    if (columnInfo.scaleType === LOG) {
        correctionFactor = 0;
        if (defined(axisObj.logDomain)) {
            extremesOperators.push(axisObj.logDomain);
            domainOperators.logDomain = [axisObj.logDomain];
            /*
        Baseline and DefaultBaseValue operators
     */
            if (defaultOperator.has('baseLine')) {
                //saveBaseline FEATURES_LOGAXIS_29
                domainOperators.updateBaseValue = {
                    definedMin: domainOperators.min,
                    basevalue: processor.baseline[index],
                    stacked,
                    rangeCheck: false
                };
            }

            if (defaultOperator.has('logConstant')) {
                domainOperators.logConstant = {
                    allowNonPositiveTicks: axisConfig.allowNonPositiveTicks
                };
            }
            return domainOperators;
        }
    }

    // [**MIGRATION**] zoom --> brush
    // #ZC2016 $index
    //Brush extremes operator
    if (defaultOperator.has('brushExtremes')) {
        if (zoomed) {
            var zoomed = chart.systemConf.chart.brush[encodeType][index],
                zoomedExtent = [zoomed.minRange, zoomed.maxRange];
            extremesOperators.push(zoomedExtent);
            domainOperators.extremes = extremesOperators;
            //saveBaseline
            domainOperators.updateBaseValue = {
                definedMin: domainOperators.min,
                basevalue: processor.baseline[index],
                stacked,
                rangeCheck: true
            };
            return domainOperators;
        }
    }

    domainOperators.baseline = baseline[index] = getPropVal(axisConfig, 'baseline.show') === true ? axisConfig.baseline.value : null;

    if (customExtent) {
        extremesOperators.push(customExtent);
        domainOperators.extremes = extremesOperators;
    } else if (savedRange) {
        extremesOperators.push(savedRange);
        domainOperators.extremes = extremesOperators;
    } else {
        var temp;
        //Waterfall processor operator
        if (defaultOperator.has('waterfallProcessor') && processor.isWaterfallEnabled) {
            extremesOperators.push({
                _type: 'waterfallProcessor'
            });
        }
        //StackExtremes operator
        else if (defaultOperator.has('stackExtremes') && stacked) {
            var stackingIndices = processor.getStackingIndices(index);
            if (stackingIndices.stackIndices.length) {
                extremesOperators.push({
                    _type: 'stackExtremes',
                    stackIndices: stackingIndices.stackIndices,
                    data: processor.encodes.stack
                });
            }
            if (stackingIndices.nonStackseriesIndices.length) {
                extremesOperators.push({
                    _type: 'extremesByAccessor',
                    indices: stackingIndices.nonStackseriesIndices,
                    data: stackingIndices._nonStackData,
                    accessor: function(encodeType, seriesIdx) {
                        let yaxiscolumnorder = DataProcessor.helpers.getParsedYAxisOrder(
                            processor.processedseriesData[seriesIdx].yaxiscolumnorder
                        );
                        return function(d) {
                            var yVal = encodeType == 'y' ? chart.dataset.getY(d, yaxiscolumnorder) : chart.dataset.getX(d);
                            return yVal;
                        };
                    },
                    isRangeType: function(seriesIdx) {
                        return chartcategory.rangePlot.has(processor.allChartTypes[seriesIdx]);
                    }
                });
            }
        } else {
            if (encodeType == 'x') {
                extremesOperators.push({
                    _type: 'extremesByAccessor',
                    data: processor.processedseriesData,
                    accessor: function(d) {
                        return function(d) {
                            return chart.dataset.getX(d);
                        };
                    }
                });
            } else {
                extremesOperators.push({
                    _type: 'extremesByEncode',
                    seriesExtermes: processor.seriesExtermes
                });
            }
        }

        if (1) {
            /*
            ThresholdExtremes
             */
            if (defaultOperator.has('threshold')) {
                let axisDataKey = encodeType == 'x' ? `chart.axes.${encodeType}axis` : `chart.axes.${encodeType}axis[${index}]`,
                    _axisConfig = pick(getPropVal(chart.userdata, axisDataKey), {}),
                    thresholdData = getThresholdValues(_axisConfig);
                if (thresholdData && thresholdData.length) {
                    extremesOperators.push(thresholdData);
                }
            }

            /*
            LevelMarker Extremes
             */
            if (defaultOperator.has('levelMarker')) {
                if (DataProcessor.helpers.hasLevelMarkerData(processor, index)) {
                    let levelMarkerIndices = processor.getLevelMarkerIndices && processor.getLevelMarkerIndices(index);
                    if (levelMarkerIndices) {
                        if (levelMarkerIndices.indices.length) {
                            extremesOperators.push({
                                _type: 'extremesByData',
                                data: processor.getEncode('y', index).seriesData,
                                indices: levelMarkerIndices.indices
                            });
                            extremesOperators.push([0, 0]);
                        }
                        if (levelMarkerIndices.values.length || processor.hasLevelMarker) {
                            extremesOperators.push(levelMarkerIndices.values);
                            extremesOperators.push([0, 0]);
                        }
                    }
                }
            }
        }

        domainOperators.extremes = extremesOperators;
    }
    domainOperators.min = axisConfig.minRange;
    domainOperators.max = axisConfig.maxRange;

    /*
        Baseline and DefaultBaseValue operators
     */
    if (defaultOperator.has('baseLine')) {
        var hasDefaultBase = dataType == NUMERIC_DATATYPE && _hasDefaultBase(encodeData, processor, chart);
        domainOperators.hasDefaultBase = hasDefaultBase;

        //saveBaseline
        domainOperators.updateBaseValue = {
            definedMin: domainOperators.min,
            basevalue: processor.baseline[index],
            stacked,
            rangeCheck: false
        };

        //If extentY = [undefined, undefined], don't change the extent with '0'
        //update baseMin Value
        //hasDefaultMin
        domainOperators.hasDefaultBaseValue = {};
    }

    //Correction factors

    /*
    Correction factor
    For time axis, correction factor also applied after extremes
     */
    if (defaultOperator.has('correctionFactor') && dataType == TIME_DATATYPE) {
        domainOperators.correctionFactor1 = {
            _needCorrectionFactor:
                encodeType == 'x' || !(processor.isPercentChart || processor.hasLevelMarker || processor.isWaterfallEnabled),
            correction: 0 //outerpadding 0 for time datatype in minmaxy file
        };
    }

    if (defaultOperator.has('subfunction')) {
        domainOperators.subfunction = {
            subfunction: subfunction
        };
    }

    if (defaultOperator.has('basevalue')) {
        if (!chart.systemConf.chart.axes[`${encodeType}axis`].disableBaseValue && !zoomed && encodeType != 'clr') {
            domainOperators.basevalue = {
                metaConf: metaConf
            };
        }
    }

    /*
        Bubble bound operators
     */
    if (defaultOperator.has('bubbleBound')) {
        if (processor.isBubbleEnabled && isUndefined(axisConfig.minRange) && isUndefined(axisConfig.maxRange)) {
            let operator = getCorrectionParser('bubble', encodeType, index, chart);
            correctionOperators.push(...operator);
            modifyCorrectionOperator('bubble', operator, domainOperators);
        }
    }

    /*
        MarkerBound operator
     */
    if (defaultOperator.has('markerBound')) {
        let operator = getCorrectionParser('marker', encodeType, index, chart);
        correctionOperators.push(...operator);
        modifyCorrectionOperator('marker', operator, domainOperators);
    }

    /*
    Correction factor
    For y axis, correction factor applied before to font bound
     */
    if (defaultOperator.has('extraCorrectionFactor') && dataType == TIME_DATATYPE) {
        domainOperators.extraCorrectionFactor = {
            dataType: dataType,
            _needCorrectionFactor: !(processor.isPercentChart || processor.hasLevelMarker || processor.isWaterfallEnabled),
            correction: correctionFactor
        };
    }

    /*
    Correction factor
    For y axis, correction factor applied before to font bound
     */
    if (defaultOperator.has('correctionFactor') && encodeType == 'y' && dataType !== TIME_DATATYPE) {
        domainOperators.correctionFactor = {
            dataType: dataType,
            _needCorrectionFactor: !(processor.isPercentChart || processor.hasLevelMarker || processor.isWaterfallEnabled),
            correction: correctionFactor
        };
    }

    // no need to find correction datalabel,if correction factor is 0
    /*
    FontBound Operator
     */
    if (defaultOperator.has('fontBound')) {
        let operator = getCorrectionParser('label', encodeType, index, chart);
        correctionOperators.push(...operator);
        modifyCorrectionOperator('label', operator, domainOperators);
        // condition missing TODO
    }

    //for time datatype, precalculation calculated from outerpaddingCorrection and fontbound
    if (defaultOperator.has('correctionFactor') && dataType == TIME_DATATYPE) {
        domainOperators.correctionFactor = {
            _needCorrectionFactor: false,
            preCorrectionValue: true,
            fontPreCorrection: true
        };
    }
    /*
    Correction factor
    For x axis, correction factor applied next to font bound
     */
    if (defaultOperator.has('correctionFactor') && encodeType == 'x' && dataType !== TIME_DATATYPE) {
        domainOperators.correctionFactor = {
            dataType: dataType,
            correction: minmaxCorrection,
            _needCorrectionFactor: true
        };
    }

    /*
        Log scale
     */
    if (defaultOperator.has('logConstant')) {
        domainOperators.logConstant = {
            allowNonPositiveTicks: axisConfig.allowNonPositiveTicks
        };
    }
    /*
    Waterfall bound operator
     */
    if (defaultOperator.has('waterfallBound') && processor.isWaterfallEnabled) {
        domainOperators.waterfallBound = {
            mode: 'domain'
        };
    }

    // if (!zoomedValue && (extent || type == 'x')) {
    //     var diff = domain[1] - domain[0];
    //     var correction = diff == 0 ? chartObj.plotarea.width / 2 : diff * outerPadding;
    //     domain[0] = hasCustomMinRange ? domain[0] : domain[0] - correction;
    //     domain[1] = hasCustomMaxRange ? domain[1] : domain[1] + correction;
    // }

    //DEMO_LINE_15
    if (defaultOperator.has('timeFormat') && dataType == TIME_DATATYPE) {
        domainOperators.timeFormat = {};
    }
    /*
    Correction factor
    For x axis, correction factor applied before to font bound
     */
    if (defaultOperator.has('extraCorrectionFactor') && dataType == NUMERIC_DATATYPE) {
        domainOperators.extraCorrectionFactor = {};

        domainOperators.correctionFactor1 = {
            _needCorrectionFactor: false,
            preCorrectionValue: true,
            fontCorrectionFactor: true
        };
    }

    domainOperators._correction = _modularized.scaleCorrection ? correctionOperators : [];

    return domainOperators;
}
Registry.setComponent('axis.parser.numeric', numericAxisParser);

function getDefaultOperator(encodeType, datatype, needOnlyExtremes, needOnlyCorrection, scaleType) {
    if (datatype == TIME_DATATYPE) {
        if (encodeType == 'x') {
            return getAvailableOperatorsForTimeXAxis(scaleType, needOnlyExtremes, needOnlyCorrection);
        }
        return getAvailableOperatorsForTimeYAxis(scaleType, needOnlyExtremes, needOnlyCorrection);
    } else {
        if (encodeType == 'x') {
            return getAvailableOperatorsForXAxis(scaleType, needOnlyExtremes, needOnlyCorrection);
        }
        return getAvailableOperatorsForYAxis(scaleType, needOnlyExtremes, needOnlyCorrection);
    }
}

function getOperatorsForPop(encodeType, index, chart, excludeDatalabels, minmaxCorrection) {
    var processor = chart.dataObject,
        dataSet = chart.dataset,
        axisConfig = getAxisData(chart, encodeType, index, null, chart.userdata),
        dataType = dataSet.getDataType(encodeType, index),
        columnInfo = dataSet.getColumn(encodeType, index),
        toBePopped = [];

    if (!(processor.isBubbleEnabled && isUndefined(axisConfig.minRange) && isUndefined(axisConfig.maxRange))) {
        toBePopped.push('bubbleBound');
    }

    if (!(DataProcessor.helpers.hasMarkerCorrection() && dataType !== TIME_DATATYPE)) {
        toBePopped.push('markerBound');
    }

    let labelCheck =
        !processor.isPercentChart &&
        !(processor.isBubbleEnabled && !processor.pieWithAxis) &&
        (processor.isAxisCategory || processor.isNonAxisMultiSeriesCategory) &&
        !(axisConfig.minRange != null && axisConfig.maxRange != null);
    if (
        !(
            (minmaxCorrection !== 0 || columnInfo.scaleType === LOG) &&
            (encodeType == 'x' || (encodeType == 'y' && !excludeDatalabels)) &&
            labelCheck
        )
    ) {
        toBePopped.push('fontBound');
    }

    if (encodeType === 'y' && processor.isPercentChart && processor.isstacked()) {
        toBePopped.push('bubbleBound', 'markerBound', 'fontBound');
    }

    return toBePopped;
}
