// @zohocharts/zc/src/scale-parser/axis/ordinal-axis-parser.js
import { getAxisOrient } from '../../axis/axisutils';
import { BOTTOM_ALIGN, TOP_ALIGN } from '../../core/align-constants';
import { getCorrectionParser } from '../../scale-parser/axis/correction/correction-parser';
import { dataLabelEnabled } from '../../label/checkLabelsEnabled';
import { modifyCorrectionOperator } from '../../scale/evaluators/correction/correction-adapter';
import { POLAR_SCALETYPE } from '../../scale/scale-type-constants';
// import { getPolorRangeConfig } from '../../polaraxis/polaraxis-helpers';
import { NULL } from '../../core/zc-constants'; // TODO: Remove at end of Code modularization
import { _modularized } from '../../core/zcconf';
import { Registry } from '../../core/zc-registry';

export function parseOrdinalAxisData() {}

function parseDomain() {
    var categories = dataObj.getCategories(type, index, applyZoom);
    domain = dataObj.getDomain(categories, type, true);

    var parsedData = {
        domain: {
            categories: {
                encode: ''
            }
        }
    };
}

export function ordinalAxisRangeParser(type, index = 0, chart, needCorrections, scaleType) {
    index = type === 'y' ? index : 0; // CXZA_MULTIY_4
    var rangeOperators = {},
        correctionOperators = [];

    var orient = getAxisOrient(type, chart.axes.rotated);
    rangeOperators.plotBound = {
        dimension: orient == BOTTOM_ALIGN || orient == TOP_ALIGN ? 'width' : 'height',
        customPlotBound: chart.systemConf.chart,
        plotSize: chart.plotarea,
        defaultPlotSize: chart.plotarea,
        polarConfig: scaleType == POLAR_SCALETYPE ? Registry.getComponent('axis.polar.helpers.getPolarRangeConfig')(chart, type) : NULL,
        color: chart.imagePallete || chart.seriesColor,
        boundExtreme: 0
    };

    if (needCorrections) {
        if ((chart.datalabels.moduleExists && dataLabelEnabled(chart)) || chart.dataObject.isBubbleEnabled) {
            if (chart.dataObject.isBubbleEnabled) {
                var operator = getCorrectionParser('bubble', type, index, chart);
                correctionOperators.push(...operator);
                modifyCorrectionOperator('bubble', operator, rangeOperators);
            } else if (!1) {
                var operator = getCorrectionParser('marker', type, index, chart);
                correctionOperators.push(...operator);
                modifyCorrectionOperator('marker', operator, rangeOperators);
            } else {
                // if (!(hasFixedDatalabels(chart) && type === 'x')) {
                var operator = getCorrectionParser('label', type, index, chart);
                correctionOperators.push(...operator);
                modifyCorrectionOperator('label', operator, rangeOperators);
                // }
            }
        }
    }
    rangeOperators._correction = _modularized.scaleCorrection ? correctionOperators : [];
    return rangeOperators;
}
Registry.setComponent('axis.parser.ordinal', ordinalAxisRangeParser);
