// @zohocharts/zc/src/components/race/core/race-init.js
import { onBrushHandler } from '../handlers/scrollBrush-handler';
import { Registry } from '../../../core/zc-registry';
import { defined, invokeFunction, isArray, isUndefined, mergeJSON } from '../../../utils/zcutils';
import { RacingChart } from './race-constructor';
import { defaultRaceCompParser, raceComponentRenderer } from '../raceComponents';
import { raceSettings } from '../race-config';
import { generateIntermediateKeyFrames } from '../intermediateFrames/init';
import { getRaceScrollView } from '../handlers/continuousRace-handler';
import { onRedrawHandler } from '../handlers/redraw-handler';
import { DataProcessor } from '../../../core/data-processor';

export function initializeRace(customConf, chart) {
    var race = chart.racing,
        isReInitRace_onRedraw = race && race.started ? chart.updating && race.allowRedraw : chart.updating;
    if (!(race && race.isInitiated) || isReInitRace_onRedraw) {
        if (chart.nochartdata) {
            return; //check this change
        }
        var defaultConf = raceSettings(chart),
            conf = mergeJSON(defaultConf, customConf, true);

        if (!race) {
            var args = getRaceArgs(chart, conf, chart.updating && race.allowRedraw);
            race = chart.racing = new RacingChart(args);
            race.userRaceSettings = conf;
            race.frameIdMetaData = race.userRaceSettings.frames.metadata.column;
            if (!race.keyFramedRace) {
                race.scrollView = getRaceScrollView(chart, !!chart.scroll.brushScale.bandwidth);
            }
        } else {
            race.userRaceSettings = conf;
            race.events.onInit = conf.events.onFrameInit;
            race.events.onLoad = conf.events.onFrameLoad;
            race.events.onInterrupt = conf.events.onFrameInterrupt;
        }
        if (race.keyFramedRace) {
            race.positionArray = [];
            generateIntermediateKeyFrames.call(race, chart.userdata.keyframes);
            race.raceScale = d3
                .scaleLinear()
                .domain(race.positionArray)
                .range(d3.range(race.positionArray.length));
            race.raceLength = race.data.length;
        } else {
            //if not ordinal, we can handle via brushScale
            var data_info = getContinuousDataPositions(chart.dataObject.processedseriesData, chart),
                positionsX = data_info.x,
                arr = [...new Set(positionsX)].sort(function(a, b) {
                    return Number(a) - Number(b);
                });
            updateYminMax(data_info, arr, chart);
            if (chart.dataset.getDataType('x') != 'ordinal') {
                chart.racing.continuousRaceInfo.raceScaleX = d3
                    .scaleLinear()
                    .domain(arr)
                    .range(d3.range(arr.length));
            }
        }
        updateDuration.call(race, race.userRaceSettings, chart);
        race.compRenderer = raceComponentRenderer;

        race.componentsParserArray = defaultRaceCompParser.call(race, chart.seriesdata, race.userRaceSettings, chart);
        raceComponentRenderer.call(race, chart.seriesdata, chart, null, race.componentsParserArray);
        race.isInitiated = true;
    }
    if (race.keyFramedRace) {
        chart.renderVisiblePointsOnly = false;
    } else {
        chart.renderVisiblePointsOnly = true;
    }
    return race;
}

function getRaceArgs(chart, userConf, allowRedraw) {
    var keyFrames = chart.systemConf.keyframes,
        keyFramedRace = chart.systemConf.keyframes ? true : false;

    var utils = $ZC.utils;
    return {
        chart: chart,
        isInitiated: false,
        raceMode: 'auto',
        keyFrameInfo: {
            index: -1
        },
        continuousRaceInfo: {
            index: -1
        },
        manualQueue: [],
        keyFramedRace: keyFramedRace,
        raceLength: keyFramedRace ? keyFrames.length : getContinuousDataLength(chart),
        data: keyFramedRace ? keyFrames : chart.seriesdata,
        cType: chart.distinctChartTypes || utils.getDistinct(utils.getAllSeriesTypes(chart.userdata)),
        replay: false,
        handlers: {
            onBrush: onBrushHandler,
            onRedraw: onRedrawHandler
        },
        events: {
            onInit: userConf.events.onFrameInit,
            onLoad: userConf.events.onFrameLoad,
            onInterrupt: userConf.events.onFrameInterrupt
        },
        allowRedraw: allowRedraw
    };
}
function getContinuousDataLength(chart) {
    var _d = chart.seriesdata,
        _l = 0;
    _d.map(function(d, i) {
        _l = d.data[0].length > _l ? d.data[0].length : _l;
    });
    return _l;
}

function updateDuration(conf, chartObj) {
    if (chartObj.userdata.keyframes) {
        var actualKeyFrameLength = chartObj.userdata.keyframes.length,
            currentKeyFrameLength = this.data.length;
        this.duration = (invokeFunction(conf.duration, this) * actualKeyFrameLength - 1) / currentKeyFrameLength;
    } else {
        this.duration = invokeFunction(conf.duration, this);
    }
}

function getContinuousDataPositions(processedseriesData, chart) {
    var xPositionArray = [],
        XtoYminMaxMap = d3.map(),
        dataProcessorHelpers = DataProcessor.helpers,
        getParsedYAxisOrder = dataProcessorHelpers.getParsedYAxisOrder;
    processedseriesData.forEach(function(d, i) {
        var curdata = d.data,
            yaxiscolumnorder = getParsedYAxisOrder(d.yaxiscolumnorder);
        curdata.forEach(subgroup => {
            if (isArray(subgroup)) {
                subgroup.forEach(d => {
                    // if (chart.dataset.getDataType('x') != 'ordinal') {
                    var _x = chart.dataset.getX(d),
                        _x_pos = chart.scroll.brushScale(_x),
                        _y = chart.dataset.getY(d, yaxiscolumnorder),
                        yval = XtoYminMaxMap.get(_x_pos);

                    // if(defined(_x_pos)){
                    _x_pos = isUndefined(_x_pos) ? null : _x_pos;
                    xPositionArray.push(_x_pos);
                    if (isUndefined(yval)) {
                        XtoYminMaxMap.set(_x_pos, {
                            ymax: _y,
                            ymin: _y
                        });
                    } else {
                        if (_y > yval.ymax) {
                            XtoYminMaxMap.set(_x_pos, {
                                ymax: _y,
                                ymin: yval.ymin
                            });
                        }
                        yval = XtoYminMaxMap.get(_x_pos);
                        if (_y < yval.ymin) {
                            XtoYminMaxMap.set(_x_pos, {
                                ymax: yval.ymax,
                                ymin: _y
                            });
                        }
                    }
                    // }
                    // }
                });
            }
        });
    });
    return {
        x: xPositionArray,
        XtoYminMaxMap
    };
}
function updateYminMax(data_info, arr, chart) {
    let YmaxArray = [],
        YminArray = [];
    if (chart.systemConf.chart.axes.xaxis.categories) {
        chart.systemConf.chart.axes.xaxis.categories.map(function(d) {
            var pos = chart.scroll.brushScale(d);
            YmaxArray.push(getYMinMaxByXPos(data_info, YmaxArray, pos));
            YminArray.push(getYMinMaxByXPos(data_info, YminArray, pos, true));
        });
    } else {
        arr.map(function(pos) {
            YmaxArray.push(getYMinMaxByXPos(data_info, YmaxArray, pos));
            YminArray.push(getYMinMaxByXPos(data_info, YminArray, pos, true));
        });
    }
    chart.racing.continuousRaceInfo.yDataMaxArr = YmaxArray;
    chart.racing.continuousRaceInfo.yDataMinArr = YminArray;
}
function getYMinMaxByXPos(data_info, array, xPos, isMin) {
    var posObj = data_info.XtoYminMaxMap.get(xPos);
    if (!array.length) {
        if (isMin) {
            return posObj ? posObj.ymax : null;
        }
        return posObj ? posObj.ymax : null;
    } else {
        var prev = array[array.length - 1];
        posObj = posObj ? posObj : null;
        var curr = isMin ? (posObj ? posObj.ymin : posObj) : posObj ? posObj.ymax : posObj;
        if (isMin ? curr < prev : curr > prev) {
            return curr;
        } else {
            return prev;
        }
    }
}

Registry.setComponent('initializeRace', initializeRace);
