// @zohocharts/zc/src/components/race/core/race-constructor.js
import { Registry } from "../../../core/zc-registry";
import { invokeFunction, isUndefined, utils } from "../../../utils/zcutils";
import { continuousRaceHandler } from "../handlers/continuousRace-handler";
import { keyFramedRaceHandler } from "../handlers/keyFramedRace-handler";

export class RacingChart {
  constructor(args) {
    var {
      chart,
      isInitiated,
      raceMode,
      conf,
      keyFrameInfo,
      continuousRaceInfo,
      manualQueue,
      keyFramedRace,
      cType,
      replay,
      data,
      raceLength,
      events, 
      handlers,
      allowRedraw
    } = args;
    this.chart = chart;
    this.isInitiated = isInitiated;
    this.raceMode = raceMode;
    this.conf = conf;
    this.keyFrameInfo = keyFrameInfo;
    this.continuousRaceInfo = continuousRaceInfo;
    this.manualQueue = manualQueue;
    this.keyFramedRace = keyFramedRace;
    this.cType = cType;
    this.replay = replay;
    this.autoPlayMode = "play";
    this.data = data;
    this.raceLength = raceLength;
    this.events = events;
    this.handlers = handlers;
    this.allowRedraw = allowRedraw;
  }

  render(chart, donotCallProcessor) {
    var race = chart.racing,
      cType = race.cType,
      dataObj = chart.dataObject,
      flattenPoints = dataObj && dataObj.flattenPoints,
      pane0_0 = flattenPoints && flattenPoints.get('pane0_0');

    invokeFunction(race.events.onInit, race, chart);
    //resetting flattenpoints each time,else it gets appended each time 
    cType.map(function(type) {
      if (pane0_0) {
        var _tp = pane0_0.get(type);
        Object.keys(_tp).map(function(d, i) {
          _tp[d] = [];
        })
      }
    });
    chart.cache.brushDataDetails = {};//plot events issue
    chart.preProcessor.processTimeData();
    // if(!donotCallProcessor){
      chart.dataObject.initialize();
    // }
    if(chart.scroll){
      chart.scroll.visiblePointCount = null;
    }
    chart.viewPortExtent = null;//any default actions of scroll is not used, so we can nullify viewportextent. if not updated to null, it affects the ticks of axis, we donot update scroll on each race, viewpoortextent updated on scroll action only.
    chart.updateScale(null, 'x');
    chart.updateScale(null, 'y', null, 0);
    chart.parseScale();
    chart.plot.renderer.draw(true);
    chart.dataObject.flattenData();

  }
  transitionStopper(chart, currentStep, maxStep) {
    return new Promise(function(resolve) {
      var t = setInterval(function () {
        var race = chart.racing,
            isLegendFiltered = race && race.legendFilterInfo && race.legendFilterInfo.isLegendFiltered;
        if (
          chart.destroying ||
          currentStep >= maxStep ||
          race.autoPlayMode === "play" ||
          race.autoPlayMode === "replay" ||
          isLegendFiltered ||
          // chart.racing.mouseDownTouchStart ||
          chart.updating
        ) {
          setTimeout(function () {
            clearInterval(t);
            resolve();
          }, race && race.duration || 100);
        }
      }, 100);
    });
  }
  setData(_frame, chart) {
    var isMultiSeriesFrame = _frame && _frame.length,
      chartData = chart.seriesdata;

    if(!_frame){
      _frame = this.data[0];
    }
    var categoricalOrder =
      _frame.keys_order || (_frame[0] && _frame[0].keys_order);
    if (isMultiSeriesFrame) {
      _frame.map(function(series, index) {
        series.disabled = chartData[index].disabled;
        Object.keys(chart.seriesdata[index]).map(function (d) {
          delete chart.seriesdata[index][d];
        });

        Object.keys(series).map(function (d) {
          chart.seriesdata[index][d] = series[d];
        });

        if (categoricalOrder) {
          chart.systemConf.chart.axes.xaxis.categories = categoricalOrder;
        }
      });
    } else {
      _frame.disabled = chartData[0].disabled;
      chart.systemConf.seriesdata.chartdata[0] = _frame;
      if (categoricalOrder) {
        chart.systemConf.chart.axes.xaxis.categories = categoricalOrder;
      }
    }

    //inorder to utilize extender in racing charts.
    let _chartSpecGetter = Registry.getComponent('dataSplicer');
      if (chart.userdata.viewParser || chart.userdata._viewParser) {
        let _viewParser = Registry.getComponent('viewParser');
        let bound = {
            canvasarea: chart.canvasarea, 
            chartarea: chart.chartarea, 
            plotarea:chart.plotarea
        }
        chart.viewFlow = new _viewParser(chart.userdata.viewParser || chart.userdata, chart.userdata, bound, chart);
        _chartSpecGetter = Registry.getComponent('dataSplicer4ViewParser');
        chart.parserConfig = chart.viewFlow.init();
    }
    chart.dataSplicer = new _chartSpecGetter(chart.userdata, chart.systemConf, chart);
    if(chart.viewFlow){
      chart.viewFlow.process(chart.seriesdata);
    }
  }
  collectRaceParams(chart) {
    var _this = this,isAuto = _this.raceMode === "auto",
      scroll = chart.scroll,
      _rp = {},
      replay = _this.replay,
      _l = _this.raceLength;
    if (isAuto) {
      if (scroll) {
        var instance = scroll.brushInstance,
          _r = instance.getRange();
        _rp.currentStep = replay ? instance.getSelection()[0] : instance.getSelection()[1];
        _rp.maxStep = _r[1] - _r[0];
        _rp.stepSize = _rp.maxStep / _l;
        if(_rp.currentStep == _rp.maxStep){
          _rp.currentStep = 0;
        }
      } else {
        //incase of ranking race chart(without scroll) and continuous race chary
        _rp.maxStep = _l - 1;
      }
    }
    return _rp;
  }
  manualRace_manager(chart, _frame, _q) {
    var _this = this;
    if (!_q.length) {
      _q = _this.manualQueue = [_frame];
      _this.initiateRacing(chart, _q);
    } else {
      _this.manualQueue.push(_frame);
    }
  }
  getKeyFrame(chart, keyFrames) {
    //frmaeInfo.index - for framed auto race
    var scrollRange =
        (chart.scroll.lastScroll && chart.phase != "redraw"
          ? chart.scroll.getScrollSelection()
          : chart.scroll.initialSelection)[1],
          raceScale = chart.racing.raceScale,
          frameInfo = this.keyFrameInfo;

    if(this.raceMode === "auto" && frameInfo.index >= 0){
      frameInfo.index = frameInfo.index > keyFrames.length || frameInfo.index < 0 ? 0 : frameInfo.index;
      scrollRange = this.chart.racing.positionArray[frameInfo.index];
      //we take scrollRange in this way, to get next keyframe properly, if got via getSelection, we might get the start value of the interpolation of selection, so previous frame might be chosen.
      //frameInfo.index  - updated in keyframedracehandlers.js
    }
    var index = /*frameInfo.index =*/ Math.round(raceScale(scrollRange));
    index = index === keyFrames.length ? keyFrames.length - 1 : index;
    index = index > keyFrames.length || index < 0 ? 0 : index;
    frameInfo.index = this.raceMode === "manual" ? index : frameInfo.index;

    return keyFrames[index];

  }
  async initiateRacing(chart) {
    var _this = this,
      _rp = _this.collectRaceParams(chart),
      scroll = chart.scroll,
      _vp = scroll && scroll.visiblePointCount,
      sliderStyle = scroll && scroll.brushInstance.options.styles.slider,
      _animateSlider = utils.isObjectPropDefined(sliderStyle, "animation"),
      instance = scroll && scroll.brushInstance,
      isKeyFramedRace = _this.keyFramedRace,
      frameInfo = _this.keyFrameInfo,
      continuousRaceInfo = _this.continuousRaceInfo,
      frameData = _this.data,
      _frame,
      manualData = _this.manualQueue,
      raceDuration = _this.duration,
      brush = scroll && scroll.brushInstance,
      breakLoop; 
    chart.racing.visibleFrameStatus = !chart.racing.visibleFrameStatus ? {
      isTransitionStarted: false,
      isRenderingStarted: false
    }:chart.racing.visibleFrameStatus;

    var resolveFunct = function(resolve, reject){
      var _i = setInterval(function(){
        if(!chart.phase){
          clearInterval(_i);
          resolve();
        }
      },10);
    }
    
    while (1) {
      chart.racing.interrupt = false;
      chart.racing.visibleFrameStatus.isRenderingStarted = true;
      chart.racing.visibleFrameStatus.isTransitionStarted = false;

      var isAuto = _this.raceMode === "auto";
      _this.syncTransition = chart.container
        .transition()
        .duration(isAuto ? raceDuration : 1)
        .ease(d3.easeLinear)
      if (_animateSlider && sliderStyle.animation.enabled) {
        sliderStyle.animation.transition =  function() {
          return _this && _this.syncTransition;
        };
      }
      if (isAuto) {
        if(!this.keyFramedRace){
          chart.scroll.visiblePointCount = null;
          var xDataType = chart.dataset.getDataType("x");
          breakLoop = continuousRaceHandler.call(this, xDataType, _rp);
          //here when scroll bar needed case, not handled, if needed can be done/ scroll height can be set to 0
        }else{
          breakLoop = keyFramedRaceHandler.call(this);
        }
        _frame = isKeyFramedRace
            ? frameData[frameInfo.index < 0 ? 0 : frameInfo.index]
            : frameData;
       

        var transEnd = _this.syncTransition.end();
        await Promise.race([transEnd, _this.transitionStopper(chart, _rp.currentStep, _rp.maxStep)]);
        
        if(brush) {
          brush._raceInProgress = false;
        }
        if(chart.racing){
          chart.racing.replay = false;
          chart.racing.visibleFrameStatus.isTransitionStarted = false;
          chart.racing.visibleFrameStatus.isRenderingStarted = false;
        }
       
        if (_animateSlider) {
          sliderStyle.animation.transition = null;
        }
        //legend case
        var legendFilterInfo = _this.legendFilterInfo;
        if(legendFilterInfo){
          if(legendFilterInfo.isLegendFiltered){
            chart.racing.syncTransition = chart.container.transition().duration(1);
            legendFilterInfo.callback(legendFilterInfo.args[0], legendFilterInfo.args[1], legendFilterInfo.args[2], legendFilterInfo.args[3])
          }
          legendFilterInfo.isLegendFiltered = false;
        }
        //redrawing
        if(chart.phase === "redraw"){
          var _interval = setInterval(function () {
            // if (_this.axesRendered) {
              clearInterval(_interval);
              var _timer = setTimeout(function () {
                chart.racing.interrupt = true;
                chart.racing.allowRedraw = true;

                // chart.racing.raceScale = null;
                // frameInfo.index = -1;
                // continuousRaceInfo.index = -1;

                if(chart.scroll){
                  chart.scroll.visiblePointCount = null;
                }
                if(isAuto){//on redraw racing-02 not got reset to initial state.
                  frameInfo.index = 0;
                  _frame = frameData[frameInfo.index];
                  _this.setData(_frame, chart);
                }
                frameInfo.index = -1;
                continuousRaceInfo.index = -1;
                continuousRaceInfo.initialValue = null;
                // if (chart.racing.scrollView) {
                  chart.racing.scrollView = (chart.scroll &&
                    chart.scroll.initialSelection) || [0, 0];
                // }
                clearTimeout(_timer);
              }, 10);
            // }
          }, 1);

          await new Promise(resolveFunct);  
        }
        
        if (chart.destroying || _this.autoPlayMode === "play" /*|| chart.racing.mouseDownTouchStart*/ || chart.racing.allowRedraw) {
          // chart.racing.allowRedraw = false;
          //chart.scroll.brushInstance.parent_group.interrupt();
          invokeFunction(_this.events.onInterrupt, _this, chart);
          chart.racing.interrupt = false;
          break;
        }
        if (invokeFunction(breakLoop)) {
          chart.racing.interrupt = false;
          _this.autoPlayMode = "replay";
          invokeFunction(_this.events.onInterrupt, _this, chart);
          frameInfo.index = -1;
          continuousRaceInfo.index = -1;
          chart.racing.scrollView = null;
          if(chart.scroll){
            chart.scroll.visiblePointCount = null;
          }
          break;
        }


      } else {
        invokeFunction(_this.events.onInit, _this, chart);
        chart.racing.visibleFrameStatus.isRenderingStarted = true;
        _frame = manualData[0];
        _this.setData(_frame, chart);
        _this.render(chart);
        await _this.syncTransition.end();
        chart.racing.visibleFrameStatus.isTransitionStarted = false;
        chart.racing.visibleFrameStatus.isRenderingStarted = false;
        chart.scrolling = false;
        if (
            manualData.length >
              2 /* if > 1 is put loop will not end manualdata will alwayss be 1. */ &&
            manualData.length === _this.prevManualDataLength - 1
          ) {
            var lastData = manualData[manualData.length - 1];
            manualData.splice(0, manualData.length);
            manualData[0] = lastData;
          } else {
            _this.prevManualDataLength = manualData.length;
            manualData.shift();
          }
        if (manualData.length === 0) {
          break;  
        } 
      }
      invokeFunction(_this.events.onLoad, _this, chart);
    }
    return new Promise(function(resolve, reject){
      resolve();//for events getting triggered while racing.
    })
  }
}
