// @zohocharts/zc/src/funnel/funnelOptions.js
import { defaultTheme as options } from '../core/defaulttheme';

options.funnel = function() {
    return {
        weighted: 'height', //height | width | none
        split: false,
        canOverlap: true,
        threeDimension: true,
        outerPadding: '2%',
        stemHeight: '20%',
        stemWidth: '20%',
        stemRadius: '0',
        borderRadius: '0 0 0 0',
        coneWidth: '100%',
        strokeColor: 'transparent',
        emptyFunnel: {
            show: false,
            color: '#FAFAFA'
        },
        datalabels: {
            showAs: 'x', //x,y,percent
            type: 'strokeddoubleside', // singleside || doubleside || strokeddoubleside
            padding: 30,
            line: {
                strokeColor: 'grey',
                strokeWidth: 0.3
            },
            innerLabel: {
                show: false,
                showAs: 'percent' //x,y,percent
            }
        },
        gradients: {
            // linear || none
            options: {
                linear: {
                    x2: 50,
                    spreadMethod: 'reflect', // pad || repeat || reflect
                    colorGamma: [0.6, 0]
                }
            }
        },
        animation: {
            type: 'vertical' //horizontalAll, verticalAll, vertical
        }
    };
};

export var _funnelOptions = options.funnel;
