// @zohocharts/zc/src/funnel/funnelchart.js
//$Id$
//funnelchart.js
import { Registry } from '../core/zc-registry';
import { legend_helpers as L_H, svg_plot_renderer } from '../core/main-zc-init';
import { utils, pick, validateNullVal, simpleClone, invokeFunction, defined, perToPx, isUndefined, pInt} from '../utils/zcutils';
import { getStrokeColor, getFillColor, hasColorColumnInNonAxis } from '../utils/colorutils';
import { getAnimationConf } from '../utils/animationutils';
import { FUNNEL_CHARTNAME } from '../core/chartname-constants';
import { NONE, mathMax, mathAbs, mathMin, POINTER } from '../core/zc-constants';
import { HORIZONTALALL_ANIMATION, VERTICAL_ANIMATION, VERTICALALL_ANIMATION } from '../core/animation-constants';
import { d3_scaleLinear, d3_ease_linear, d3_interpolate, d3_select, d3_event, d3_line } from '../core/d3-constants';
import { appendEle } from '../utils/zcsvgutils';
import { PATH_ELEMENT } from '../core/element-constants';
import { setPointerEvents, getEventObject } from '../events/events-utils';
import { generateDataForAnnotation } from '../annotation/annotation-utils';
import { labelRenderer } from '../core/main-zc-init';
import { attachevents } from '../events/eventActions';
import { effects } from '../effects/zceffects';
import { isDatalabelEnabled } from '../label/datalabelsUtils';
import { getEventConf } from '../events/events-utils';
// import { legend_helpers as L_H } from '../legend/legend-helper';
import { updateSelectionOnDatapoint } from '../events/event-update-selection';
import { isFacet } from '../axis/hierarchy-axis-utils';
import { setOpacityForTickFilterElements, setPivotAttributes } from '../plot/svg/set-pivot-attr';
import { zcMorph } from '../morphing/zc-morping';
import { getRawColor } from '../utils/colorutils';
import { getOffsetValue } from '../plot/morph/morphTools';

let annotation = Registry.getComponent('annotation');
/**
 *
 * @returns {chart} funnel renderer.
 */
svg_plot_renderer.funnel = function () {
    var MAXWIDTH_DATALABELS = 0.3;
    var cfg = {};
    var commonRendererProp,
        chartObj,
        data_processed,
        noofNonZeroDatas,
        coneHeigth,
        noOfGaps,
        stemHeight,
        stemWidth,
        xAxisWidth,
        dataset,
        percentage = 0.375,
        chartWidth,
        totalEngagement,
        trapezoid,
        chartHeight,
        color,
        funnelsections,
        ellipseSections;
    var threeDimension,
        arcHeight,
        arcHeight1,
        padding,
        minPadding,
        maxPadding,
        minArcHeight = 1,
        isredraw = false,
        arcCorrection = 0,
        container,
        ellipse,
        funnel,
        label,
        labelLinkgroup,
        rightXscale,
        leftXscale,
        yaxiscolumnorder;
    var plotoptions,
        commonPlotOptions,
        systemconf,
        notesArray,
        datalabelPadding,
        isSplit,
        outerPadding,
        gradientDim,
        isPatternPalleteEnabled;
    var maxPer,
        isWeightedWidth,
        noWeighted,
        defsLocation,
        rp,
        gElementsCollected,
        customPlotarea,
        fillColor = [],
        paneid;
    let shiftInfo;

    function chart(selection) {
        selection.each(function (data) {
            //renderer config values
            rp = commonRendererProp.params;
            (chartObj = rp.chartObj),
                (commonPlotOptions = rp.commonPlotOptions),
                (color = rp.color),
                (notesArray = rp.notesArray),
                (yaxiscolumnorder = rp.yaxiscolumnorder),
                (dataset = rp.dataset);
            systemconf = chartObj.systemConf;
            var labelQueue = rp.labelQueue,
                { bound, rendererIndex } = rp,
                seriesData = isFacet(chartObj) && hasColorColumnInNonAxis(chartObj) ? data : data[0];
            paneid = isFacet(chartObj) ? `${seriesData.row}_${seriesData.column}` : null;

            defsLocation = chartObj.defsLocation;

            let plotarea = { ...bound, ...customPlotarea };
            chartWidth = plotarea.width;
            chartHeight = plotarea.height;

            gElementsCollected = false;

            plotoptions = data[0].plotoptions;
            var gvp = utils.getValueOfPlotOptions(plotoptions, commonPlotOptions);
            isPatternPalleteEnabled = systemconf.legend.imagePallete || L_H.hasPatternPallete(chartObj);
            cfg.canOverlap = gvp('canOverlap');
            isSplit = gvp('split');
            threeDimension = gvp('threeDimension');
            var emptyFunnelOptions = commonPlotOptions.emptyFunnel;
            datalabelPadding = commonPlotOptions.datalabels.padding;
            var showlabel = chartObj.datalabels.moduleExists && isDatalabelEnabled(commonPlotOptions.datalabels, chartObj);

            var strokeColor = getStrokeColor(chartObj, 0);

            minPadding = pick(commonPlotOptions.minPadding, threeDimension ? 4 : 1);
            maxPadding = pick(commonPlotOptions.maxPadding, 10);

            //Animation variables
            var animationOptions = getAnimationConf(commonPlotOptions, plotoptions, chartObj, FUNNEL_CHARTNAME);
            var animationType = animationOptions.type,
                easingType = animationOptions.easingType,
                isAnimationEnable = animationOptions.enabled,
                transitionDuration = animationOptions.duration,
                redrawDuration = isAnimationEnable ? 200 : 0;

            outerPadding = perToPx(gvp('outerPadding'), chartHeight, 0) / 2;
            stemHeight = gvp('stemHeight');
            stemWidth = gvp('stemWidth');
            let borderRadius = gvp('borderRadius');
            let stemRadius = gvp('stemRadius');
            const border = {};
            (isWeightedWidth = gvp('weighted') === 'width'), (noWeighted = gvp('weighted') === NONE);

            if (isWeightedWidth) {
                stemHeight = stemWidth = 0;
                isSplit = true;
            }

            if (borderRadius !== null) {
                if ( /* !(stemWidth == 0 || stemWidth == '0%') || */ !(stemHeight == 0 || stemHeight == '0%')) {
                    borderRadius = null;
                    stemRadius = null;
                } else if (typeof borderRadius === 'number') {
                    border.topLeft = border.topRight = border.bottomLeft = border.bottomRight = borderRadius;
                }
                else {
                    if (Array.isArray(borderRadius) || typeof borderRadius === 'string') {
                        const _borderRadius = (typeof borderRadius === 'string') ? borderRadius.split(' ').map(Number) : borderRadius;
                        border.bottomLeft = _borderRadius[0];
                        border.bottomRight = _borderRadius[1];
                        border.topRight = _borderRadius[2];
                        border.topLeft = _borderRadius[3];
                    } else if (typeof borderRadius === 'object') {
                        border.bottomLeft = borderRadius.bottomLeft;
                        border.bottomRight = borderRadius.bottomRight;
                        border.topRight = borderRadius.topRight;
                        border.topLeft = borderRadius.topRight;
                    }
                }
            }

            data_processed = data = hasColorColumnInNonAxis(chartObj) ? data : data[0].data[0];
            noofNonZeroDatas = data_processed.reduce(function (prev, cur, i) {
                return prev + (dataset.getY(cur, yaxiscolumnorder) && !cur.disabled ? 1 : 0);
            }, 0);
            noOfGaps = noofNonZeroDatas - 1;
            var preArcHeight = preCalArcHeight();
            cfg.height = chartHeight - 2 * preArcHeight - chartHeight * 0.01 - 2 * outerPadding;

            var coneWidth = gvp('coneWidth');
            coneWidth = perToPx(coneWidth, cfg.height, cfg.height * 0.95);
            cfg.width = mathMin(chartWidth, coneWidth);

            padding = getFunnelPadding(cfg.height);
            cfg.gapHeight = !isAnimationEnable || isSplit || isredraw ? padding : animationType == HORIZONTALALL_ANIMATION ? padding : 0;

            if ((isWeightedWidth || noWeighted) && defined(commonPlotOptions.maxSliceHeight)) {
                /*
                 * Set max height for each trapezoid if isWeightedWidth or noWeighted
                 */
                cfg.height -= mathMax(
                    0,
                    cfg.height - padding * noOfGaps - perToPx(commonPlotOptions.maxSliceHeight, cfg.height) * noofNonZeroDatas
                );
            }
            if (showlabel) {
                /*
                 *If chart width is lesser than height, datalabels are hidden as default padding is 30 and there is no space to show datalabels.
                 * So 30% of the width is allocated for datalabels and the default padding will be changed to 5
                 */
                var minDatalabelsWidth = chartWidth * MAXWIDTH_DATALABELS,
                    availableDatalabelsWidth = chartWidth - cfg.width;
                if (availableDatalabelsWidth < minDatalabelsWidth) {
                    datalabelPadding = cfg.datalabelPadding = 5;
                    cfg.width -= mathMax(minDatalabelsWidth - availableDatalabelsWidth, 0);
                }
            }
            stemHeight = perToPx(stemHeight, cfg.height, cfg.height * 0.2);
            // borderRadius = perToPx(borderRadius, cfg.height, cfg.height * 0.2);
            xAxisWidth = stemWidth = perToPx(stemWidth, cfg.width, (stemHeight * cfg.width) / cfg.height) / 2;
            coneHeigth = cfg.height - stemHeight;
            arcHeight = cfg.height * 0.05;

            arcHeight = mathMax(minArcHeight, (arcHeight - data_processed.length) / data_processed.length);
            arcHeight1 = arcHeight = !threeDimension ? 0 : cfg.canOverlap ? arcHeight + maxPadding / 2 : padding / 2 - arcCorrection;
            chart.total = totalEngagement = data_processed.reduce(function (prev, cur, i) {
                return prev + (data_processed[i].disabled ? 0 : mathAbs(dataset.getY(data_processed[i], yaxiscolumnorder) || 0));
            }, 0);
            maxPer = mathMax.apply(
                null,
                data_processed.map(function (d, i) {
                    return data_processed[i].disabled ? 0 : mathAbs(dataset.getY(data_processed[i], yaxiscolumnorder)) / totalEngagement;
                })
            );
            leftXscale = d3_scaleLinear()
                .domain(isWeightedWidth ? [0, maxPer] : [stemHeight, cfg.height])
                .range([cfg.width / 2 - stemWidth, 0]);
            rightXscale = d3_scaleLinear()
                .domain(isWeightedWidth ? [0, maxPer] : [stemHeight, cfg.height])
                .range([cfg.width / 2 + stemWidth, cfg.width]);
            trapezoid = funnelLayout(data_processed, { border: border, stemRadius: stemRadius });
            gradientDim = [cfg.width, cfg.height];
            chart.fillColorFunc = getFillColor(chartObj, 0, null, gradientDim);

            if (isFacet(chartObj)) {
                data_processed.forEach((d, i) => {
                    fillColor[i] = getFillColor(chartObj, i, paneid, gradientDim);
                });
            } else {
                fillColor[0] = chart.fillColorFunc;
            }

            var eventDisabled = validateNullVal(getEventConf(plotoptions, commonPlotOptions, chartObj, 'enabled'), true) == false;
            var cursor = eventDisabled ? null : getEventConf(plotoptions, commonPlotOptions, chartObj, 'cursor') || POINTER;

            let transX = perToPx(invokeFunction(gvp('x'), chartObj, cfg), chartWidth, chartWidth / 2 - cfg.width / 2),
                transY = perToPx(invokeFunction(gvp('y'), chartObj, cfg), chartHeight, arcHeight + 2 + outerPadding);

            (chartObj.plotarea.centerX = transX), (chartObj.plotarea.centerY = transY);

            var containerArea = d3_select(this);
            container = appendEle(containerArea, 'g', [data], '.', 'class', 'funnelgroup');
            container.attr('transform', `translate(${transX},${transY})`).styles({
                cursor: cursor,
                'stroke-opacity': gvp('strokeOpacity'),
                'stroke-width': gvp('strokeWidth')
            });
            ellipse = appendEle(container, 'g', [1], '.', 'class', 'ellipsegroup');
            funnel = appendEle(container, 'g', [1], '.', 'class', 'trapezoidgroup');

            funnelsections = appendEle(funnel, PATH_ELEMENT, data_processed);
            changeSelectionOrder(funnelsections);

            funnelsections
                .attrs({
                    class: 'zc_funnel',
                    id: (d, i) => `funnel${i}`
                })
                .styles({
                    'fill-opacity': commonPlotOptions.fillOpacity,
                    fill: function (d, i) {
                        if (isFacet(chartObj)) {
                            return fillColor[i](i, d);
                        }

                        return fillColor[0](i, d);
                    },
                    stroke: function (d, i) {
                        if (isFacet(chartObj)) {
                            return getStrokeColor(chartObj, i, paneid)(i, d, commonPlotOptions.strokeColor);
                        }

                        return strokeColor(i, d, commonPlotOptions.strokeColor);
                    },
                    'stroke-width': (d) => { return (d && d.disabled) ? 0 : null }
                });

            ellipseSections = appendEle(ellipse, PATH_ELEMENT, threeDimension ? data_processed : []);
            changeSelectionOrder(ellipseSections);

            ellipseSections
                .attrs({
                    class: 'zc_funnel',
                    id: (d, i) => `ellipse${i}`
                })
                .styles({
                    'fill-opacity': commonPlotOptions.fillOpacity,
                    fill: function (d, i) {
                        if (isFacet(chartObj)) {
                            return fillColor[i](i, d, 0.5);
                        }

                        return fillColor[0](i, d, 0.5);
                    },
                    stroke: function (d, i) {
                        if (isFacet(chartObj)) {
                            return getStrokeColor(chartObj, i, paneid)(i, d, commonPlotOptions.strokeColor, 0.5);
                        }

                        return strokeColor(i, d, commonPlotOptions.strokeColor, 0.5);
                    }
                });

            setPivotAttributes(funnelsections, chartObj);
            setPivotAttributes(ellipseSections, chartObj);

            setOpacityForTickFilterElements(funnelsections, chartObj, 'funnel');
            setOpacityForTickFilterElements(ellipseSections, chartObj, 'funnel');

            setPointerEvents(chartObj, funnelsections, eventDisabled, paneid);
            setPointerEvents(chartObj, ellipseSections, eventDisabled, paneid);
            if (!chartObj.cache.search) {
                chartObj.cache.search = {};
            }
            const hasChartEvents = !Registry.getComponent('ZC_ChartEvents')._empty;
            if (!eventDisabled && hasChartEvents) {
                var mousemove = function (e, d) {
                    var args = getEventArgs(e, d);
                    chartObj.eventHandler.handlers.mousemove.call(chartObj.eventHandler, e, args);
                    updateSelectionOnDatapoint(d, args);
                    chartObj.cache.search.lastPane = paneid;
                },
                    mouseout = function (e, d) {
                        var args = getEventArgs(e, d);
                        chartObj.eventHandler.handlers.mouseout.call(chartObj.eventHandler, e, args);
                        updateSelectionOnDatapoint(d, args);
                        chartObj.cache.search.lastPane = null;
                    },
                    click = function (e, d) {
                        var args = getEventArgs(e, d);
                        chartObj.eventHandler.handlers.click.call(chartObj.eventHandler, e, args);
                        updateSelectionOnDatapoint(d, args);
                        chartObj.cache.search.lastPane = paneid;
                    },
                    dblclick = function (e, d) {
                        var dt = !annotation._empty ? generateDataForAnnotation(d, null) : d;
                        var coordinates = trapezoid[data.length - 1 - d.index].bottomLeft;
                        coordinates = [coordinates[0] + (chartWidth / 2 - cfg.width / 2), coordinates[1] + (arcHeight + 2 + outerPadding)];
                        var args = getEventArgs(e, d);
                        args.coordinate = coordinates;
                        args.basedOn = 'plot';
                        args.data = dt;
                        chartObj.eventHandler.handlers.doubleclick.call(chartObj.eventHandler, e, args);
                        updateSelectionOnDatapoint(d, args);
                        chartObj.cache.search.lastPane = paneid;
                    },
                    eventAttacherOptions = {
                        target: 'plot',
                        touchleave: true,
                        chartObj: chartObj
                    };
                var mouseHandlers = {
                    mousemove: mousemove,
                    mouseout: mouseout,
                    click: click,
                    dblclick: dblclick
                },
                    touchHandlers = {
                        singletouchmove: mousemove,
                        touchend: mouseout,
                        tap: click,
                        dbltap: dblclick
                    };
                attachevents(funnelsections, mouseHandlers, touchHandlers, eventAttacherOptions);
                attachevents(ellipseSections, mouseHandlers, touchHandlers, eventAttacherOptions);
            }

            var A = cfg.canOverlap ? ((2 * stemWidth) / cfg.width) * arcHeight : arcHeight + arcCorrection;
            var funnelOutline =
                `M0,0L${cfg.width / 2 - stemWidth},${coneHeigth}L${cfg.width / 2 - stemWidth},${cfg.height
                }A${stemWidth},${A},0,0,0,${cfg.width / 2 + stemWidth},${cfg.height}L${cfg.width / 2 + stemWidth},${coneHeigth}L${cfg.width
                },0A${cfg.width / 2},${arcHeight},0,0,0,0,0` + `A${cfg.width / 2},${arcHeight},0,0,0,${cfg.width},0`;
            var eFunnel = appendEle(
                container,
                PATH_ELEMENT,
                !isredraw && emptyFunnelOptions.show && isAnimationEnable ? [1] : [],
                '#',
                'id',
                'emptyFunnel'
            );
            eFunnel.attr('d', funnelOutline).styles({
                stroke: emptyFunnelOptions.color || 'grey',
                fill: NONE
            });
            if (!(chartObj.systemConf.chart.plot.morph.enabled && chartObj.oldSeriesInfo != undefined)) {
                addFunnelClipPath(funnelOutline);
            }
            if (!isredraw) {
                hideDisabledElements(data, 0);
                var prev = 0,
                    funneldrawn = 0;

                var index = 0;
                if (chartObj.systemConf.chart.plot.morph.enabled && chartObj.oldSeriesInfo != undefined) {
                    var morph_Obj = new zcMorph({ sampling: chartObj.systemConf.chart.plot.morph.samplingDistance });
                    var oldseriesData = chartObj.oldSeriesInfo.get(index);
                    var oldseriesIndex = oldseriesData.seriesIndex;
                    var oldseriesColor = oldseriesData.color;
                    var offsetValue = getOffsetValue(chartObj, oldseriesIndex);
                    var morphDuration = chartObj.systemConf.chart.plot.morph.duration;
                    var individualElementTransform = oldseriesData.transformationInfo;
                    var line = d3_line();
                    var interpolated = {};

                    funnelsections.each(function (d, i) {
                        var fromPath = chartObj.oldSeriesInfo.get(oldseriesIndex).pathData[i];
                        var fromPathAttribute = fromPath.getAttribute('d');
                        fromPath.setAttribute('d', drawTrapezoid(trapezoid[data_processed.length - 1 - i]));
                        var toPath = fromPath.cloneNode();
                        fromPath.setAttribute('d', fromPathAttribute);
                        interpolated[i] = morph_Obj.interpolate(
                            {
                                path: fromPath,
                                xOffset: offsetValue.x + individualElementTransform[i].x,
                                yOffset: offsetValue.y + individualElementTransform[i].y
                            },
                            { path: toPath, xOffset: 0, yOffset: 0 }
                        );
                    });
                    funnelsections
                        .style('fill', function (d, i) {
                            return oldseriesColor[i];
                        })
                        .transition()
                        .duration(function () {
                            return morphDuration;
                        })
                        .ease(easingType)
                        .style('fill', function (d, i) {
                            return getRawColor(chartObj, i, d.data, oldseriesIndex, null);
                        })
                        .attrTween('d', function (d, i) {
                            var interpolator = interpolated[i];
                            return function (t) {
                                return line(
                                    interpolator.map(function (p) {
                                        return p(t);
                                    })
                                );
                            };
                        })
                        .on('end', function (d, i) {
                            if (chartObj.oldSeriesInfo) {
                                delete chartObj.oldSeriesInfo;
                            }
                            d3_select(this).style('fill', function () {
                                return (isFacet(chartObj) ? fillColor[i] : fillColor[0])(i, d, null);
                            });
                            funneldrawn++;
                            this.setAttribute('d', drawTrapezoid(trapezoid[data_processed.length - 1 - i]));
                        });
                } else {
                    funnelsections
                        .transition()
                        .duration(function (d, i) {
                            var yVal = dataset.getY(d, yaxiscolumnorder);
                            return animationType == VERTICAL_ANIMATION
                                ? d.disabled || !yVal
                                    ? 0
                                    : isWeightedWidth || noWeighted
                                        ? transitionDuration / noofNonZeroDatas
                                        : (mathAbs(yVal) / totalEngagement) * transitionDuration
                                : transitionDuration;
                        })
                        .delay(function (d, i) {
                            var yVal = dataset.getY(d, yaxiscolumnorder);
                            prev +=
                                d.disabled || !yVal
                                    ? 0
                                    : isWeightedWidth || noWeighted
                                        ? transitionDuration / noofNonZeroDatas
                                        : mathAbs(yVal);
                            var del =
                                animationType == VERTICAL_ANIMATION
                                    ? isWeightedWidth || noWeighted
                                        ? transitionDuration - prev
                                        : ((totalEngagement - prev) / totalEngagement) * transitionDuration
                                    : 0;
                            return del || 0;
                        })
                        .ease(easingType)
                        .attrTween('d', function (d, i) {
                            return tweenAll(d, data_processed.length - 1 - i, 'funnel', animationType);
                        })
                        .on('end', function (d, i) {
                            funneldrawn++;
                        });
                    prev = 0;
                    ellipseSections
                        .transition()
                        .duration(function (d, i) {
                            var yVal = dataset.getY(d, yaxiscolumnorder);
                            return animationType == VERTICAL_ANIMATION
                                ? d.disabled || !yVal
                                    ? 0
                                    : isWeightedWidth || noWeighted
                                        ? transitionDuration / noofNonZeroDatas
                                        : (mathAbs(yVal) / totalEngagement) * transitionDuration
                                : transitionDuration;
                        })
                        .delay(function (d, i) {
                            var yVal = dataset.getY(d, yaxiscolumnorder);
                            prev +=
                                d.disabled || !yVal
                                    ? 0
                                    : isWeightedWidth || noWeighted
                                        ? transitionDuration / noofNonZeroDatas
                                        : mathAbs(yVal);
                            var del =
                                animationType == VERTICAL_ANIMATION
                                    ? isWeightedWidth || noWeighted
                                        ? transitionDuration - prev
                                        : ((totalEngagement - prev) / totalEngagement) * transitionDuration
                                    : 0;
                            return del || 0;
                        })
                        .ease(easingType)
                        .attrTween('d', function (d, i) {
                            return tweenAll(d, data_processed.length - 1 - i, 'ellipse', animationType);
                        });
                }
                var includeGap = setInterval(function () {
                    if (funneldrawn == data_processed.length) {
                        if (animationType != HORIZONTALALL_ANIMATION) {
                            cfg.gapHeight = padding;
                            trapezoid = funnelLayout(data_processed, { border: border, stemRadius: stemRadius });
                            funnelsections
                                .transition()
                                .duration(isAnimationEnable ? 250 : 0)
                                .delay(50)
                                .ease(d3_ease_linear)
                                .attr('d', function (d, i) {
                                    return drawTrapezoid(trapezoid[data_processed.length - 1 - i]);
                                });
                            ellipseSections
                                .transition()
                                .duration(isAnimationEnable ? 250 : 0)
                                .delay(50)
                                .ease(d3_ease_linear)
                                .attr('d', function (d, i) {
                                    return drawEllipse(trapezoid[data_processed.length - 1 - i]);
                                });
                        }
                        lableManagent();
                        container
                            .selectAll(`${PATH_ELEMENT}#emptyFunnel`)
                            .transition()
                            .duration(50)
                            .delay(310)
                            .style('stroke-opacity', 0)
                            .remove();
                        clearInterval(includeGap);
                    }
                }, 10);
            } else {
                funnelsections
                    .transition()
                    .duration(redrawDuration)
                    .ease(d3_ease_linear)
                    .attr('d', function (d, i) {
                        return drawTrapezoid(trapezoid[data_processed.length - 1 - i]);
                    });
                ellipseSections
                    .transition()
                    .duration(redrawDuration)
                    .ease(d3_ease_linear)
                    .attr('d', function (d, i) {
                        return drawEllipse(trapezoid[data_processed.length - 1 - i]);
                    });
                hideDisabledElements(data, redrawDuration);
                lableManagent();
            }

            function lableManagent() {
                chart.renderedTimer = setTimeout(
                    function () {
                        var labelPosition = calculatePointsForLabelPlacement(data_processed, trapezoid);
                        if (labelRenderer) {
                            var translation = hasColorColumnInNonAxis(chartObj)
                                ? [transX + plotarea.translate[0], transY + plotarea.translate[1]]
                                : [transX, transY];

                            var args = [
                                chartObj.seriesdata,
                                0,
                                data_processed,
                                noofNonZeroDatas,
                                arcHeight + 2 + outerPadding,
                                totalEngagement,
                                cfg,
                                labelPosition,
                                chart.subSeriesIndex || 0,
                                translation
                            ];
                            labelQueue.push({
                                renderer: chart,
                                charttype: FUNNEL_CHARTNAME,
                                chartObj: chartObj,
                                args: args,
                                rendererIndex,
                                paneid
                            });
                            labelQueue.push({
                                renderer: chart,
                                charttype: 'funnelInnerLabels',
                                chartObj: chartObj,
                                args: args,
                                rendererIndex,
                                paneid: pick(paneid, 0)
                            });
                        }

                        //Annotation preprocessing
                        if (!annotation._empty) {
                            notesArray = annotation.preprocessAnnotations(
                                chartObj.seriesdata,
                                0,
                                function (dt, k, j) {
                                    return [
                                        trapezoid[trapezoid.length - 1 - j].bottomLeft[0] + (chartWidth / 2 - cfg.width / 2),
                                        trapezoid[trapezoid.length - 1 - j].bottomLeft[1] + (arcHeight + 2 + outerPadding)
                                    ];
                                },
                                notesArray,
                                chartObj,
                                rp
                            );
                        }

                        isredraw = true;
                        chart.chartRendered = true;

                        if (hasColorColumnInNonAxis(chartObj)) {
                            chartObj.renderer[rendererIndex].endFunctions();
                        }
                    },
                    !isAnimationEnable ? 0 : isredraw ? redrawDuration : 300
                );
                if (chartObj.timerObj) {
                    chartObj.timerObj[chartObj.timerObj.length] = chart.renderedTimer;
                }
            }

            function changeSelectionOrder(selection) {
                /*
                 * Reverse the selection order. ISSUE: https://docs.zoho.com/viewcommentattachment/4807920000000048005?width=770&height=515&owner=42061464
                 */
                selection._groups[0].reverse();
                selection.data(data_processed);
            }

            function getEventArgs(e, d) {
                var ev = getEventObject(e);
                const i = !isFacet(chartObj) ? d.index : pInt(
                    e.srcElement
                        .getAttribute('id')
                        .replace(/funnel|ellipse/gi, '')
                );
                return {
                    renderer: chart,
                    seriesIndex: hasColorColumnInNonAxis(chartObj) ? i : 0,
                    subSeriesIndex: 0,
                    itemIndex: i,
                    paneid: paneid,
                    rendererIndex: rendererIndex,
                    color: color[i % color.length],
                    coordinates: [ev.pageX, ev.pageY],
                    element: this,
                    data: d,
                    selected: d.selected,
                    _key: `ZC_${chartObj.id}_0_0_${i}`
                };
            }
        });
    }

    chart.commonRendererProp = function (_) {
        if (!arguments.length) {
            return commonRendererProp;
        }
        commonRendererProp = _;
        return chart;
    };

    chart.plotarea = function (_) {
        //Bubble Pie chart will give you the plotarea
        if (!arguments.length) {
            return customPlotarea;
        }
        customPlotarea = _;
        return chart;
    };

    function funnelLayout(data, addson = { border: null, stemRadius: null }) {
        var sumofvalues = 0,
            hgt = cfg.height,
            ht = 0,
            h = 0;
        (trapezoid = []), (xAxisWidth = stemWidth);
        // const data_length = data.length;
        data.forEach(function (value, index) {
            if (!defined(value.index)) { value.index = index; }
            // value.index = defined(value.index)  index;
            var d = data[data.length - 1 - index],
                yVal = dataset.getY(d, yaxiscolumnorder),
                disabled = d.disabled;
            h = !totalEngagement
                ? 0
                : (disabled ? 0 : isWeightedWidth || noWeighted ? (yVal ? 1 / noofNonZeroDatas : 0) : mathAbs(yVal) / totalEngagement) *
                (cfg.height - (noofNonZeroDatas - 1) * cfg.gapHeight);
            h = h || 0;
            sumofvalues += disabled ? 0 : yVal;
            h = ht + h;
            var per = yVal / totalEngagement;
            per = per || 0;
            var points = {};
            if (sumofvalues == totalEngagement && (yVal == 0 || disabled)) {
                points.topLeft = points.bottomLeft = points.middleLeft = [leftXscale(isWeightedWidth ? maxPer : cfg.height), 0];
                points.topRight = points.bottomRight = points.middleRight = [rightXscale(isWeightedWidth ? maxPer : cfg.height), 0];
            } else {
                if (isWeightedWidth) {
                    points.bottomLeft = [leftXscale(per), cfg.height - h];
                    points.topLeft = [leftXscale(per), hgt];
                    points.middleLeft = [leftXscale(per), cfg.height - h];
                    points.bottomRight = [rightXscale(per), cfg.height - h];
                    points.topRight = [rightXscale(per), hgt];
                    points.middleRight = [rightXscale(per), cfg.height - h];
                } else {
                    points.bottomLeft = [cfg.height - h >= coneHeigth ? cfg.width / 2 - stemWidth : leftXscale(h), cfg.height - h];
                    points.topLeft = [hgt >= coneHeigth ? cfg.width / 2 - stemWidth : leftXscale(ht), hgt];
                    points.middleLeft = [
                        hgt >= coneHeigth ? cfg.width / 2 - stemWidth : points.bottomLeft[0],
                        hgt >= coneHeigth && cfg.height - h <= coneHeigth ? coneHeigth : cfg.height - h
                    ];
                    points.bottomRight = [cfg.height - h >= coneHeigth ? cfg.width / 2 + stemWidth : rightXscale(h), cfg.height - h];
                    points.topRight = [hgt >= coneHeigth ? cfg.width / 2 + stemWidth : rightXscale(ht), hgt];
                    points.middleRight = [
                        hgt >= coneHeigth ? cfg.width / 2 + stemWidth : points.bottomRight[0],
                        hgt >= coneHeigth && cfg.height - h <= coneHeigth ? coneHeigth : cfg.height - h
                    ];
                }
            }
            points = { ...addson, ...points, index: index, disabled: value.disabled };
            trapezoid.push(points);
            hgt = !yVal || disabled ? cfg.height - h : cfg.height - h - cfg.gapHeight;
            ht = !yVal || disabled ? h : h + cfg.gapHeight;
        });
        return trapezoid;
    }

    function calculatePointsForLabelPlacement(data, trapezoid) {
        var trape = [],
            hgt = cfg.height,
            ht = 0,
            h = 0,
            _x = (_, isLeft) => {
                if (isLeft) {
                    return _ >= coneHeigth ? cfg.width / 2 - stemWidth : leftXscale(cfg.height - _);
                } else {
                    return _ >= coneHeigth ? cfg.width / 2 + stemWidth : rightXscale(cfg.height - _);
                }
            };
        data.forEach(function (value, index) {
            var d = data[data.length - 1 - index],
                yVal = dataset.getY(d, yaxiscolumnorder),
                disabled = d.disabled,
                point = trapezoid[index];
            h =
                ((disabled ? 0 : isWeightedWidth || noWeighted ? (yVal ? 1 / noofNonZeroDatas : 0) : mathAbs(yVal) / totalEngagement) *
                    (cfg.height - (noofNonZeroDatas - 1) * cfg.gapHeight)) /
                2;
            h = h || 0;
            var half = h;
            h = ht + h;
            var per = yVal / totalEngagement;
            per = per || 0;
            var leftX = isWeightedWidth ? leftXscale(per) : cfg.height - h >= coneHeigth ? cfg.width / 2 - stemWidth : leftXscale(h);
            var rightX = isWeightedWidth ? rightXscale(per) : cfg.height - h >= coneHeigth ? cfg.width / 2 + stemWidth : rightXscale(h);
            trape.push({
                rightx: rightX + datalabelPadding,
                leftx: leftX - datalabelPadding,
                y: mathAbs(cfg.height - h),
                y0: mathAbs(cfg.height - h + half),
                polygon: [point.bottomRight, point.middleRight, point.topRight, point.topLeft, point.middleLeft, point.bottomLeft],
                _x
            });

            h = h + half;
            hgt = !yVal || disabled ? cfg.height - h : cfg.height - h - cfg.gapHeight;
            ht = !yVal || disabled ? h : h + cfg.gapHeight;
        });
        return trape;
    }

    function addFunnelClipPath(emptyFunnelPath) {
        var clipID;
        if (isWeightedWidth) {
            clipID = null;
        } else {
            var properties = {
                name: 'FunnelClipPath',
                id: chartObj.id,
                d: emptyFunnelPath
            },
                args = {
                    chartObj: chartObj
                };
            clipID = effects.get(effects.svg_clips, 'polygon', properties, args);
        }
        container.attr('clip-path', !clipID ? null : `url(${defsLocation}#${clipID})`);
    }

    chart.defaultMouseOverFunc = function (d, seriesPos, itemPos) {
        if (!gElementsCollected) {
            collectGroups(commonRendererProp.params.rendererIndex);
        }
        //Reset previous pane mouseover
        // if(chartObj.dataObject.pivot){
        //     var allEllipseGroups = chartObj.wrap.selectAll('g.ellipsegroup'),
        //         allTrapezoidGroups = chartObj.wrap.selectAll('g.trapezoidgroup');
        //     if(isPatternPalleteEnabled){
        //         allTrapezoidGroups.selectAll('path').style('opacity', null);
        //         allEllipseGroups.style('opacity', null);
        //     }else{
        //         allTrapezoidGroups.selectAll('path').style('fill', (d,i)=>{
        //             return fillColor[i](i,d,null);
        //         });
        //         allEllipseGroups.selectAll('path').style('fill', (d,i)=>{
        //             return fillColor[i](i,d,0.5);
        //         });
        //     }
        //
        // }

        if (isPatternPalleteEnabled) {
            funnelsections.style('opacity', function (d, i) {
                return i === itemPos ? null : 0.4;
            });
            ellipseSections.style('opacity', function (d, i) {
                return i === itemPos ? null : 0.4;
            });
        } else {
            funnelsections.style('fill', function (d, i) {
                return (isFacet(chartObj) ? fillColor[i] : fillColor[0])(i, d, i === itemPos ? 0.6 : null);
            });
            ellipseSections.style('fill', function (d, i) {
                return (isFacet(chartObj) ? fillColor[i] : fillColor[0])(i, d, i === itemPos ? 1 : 0.5);
            });
        }

        let subSeriesIndex = chart.subSeriesIndex || 0;
        label.selectAll('text').style('text-shadow', null);
        labelLinkgroup.selectAll(PATH_ELEMENT).style('stroke-width', null);
        label.selectAll(`#lbl0_${subSeriesIndex}_${data_processed.length - 1 - itemPos}`).style('text-shadow', '0 0 0 rgba(0,0,0,1)');
        labelLinkgroup.selectAll(`#link0_${subSeriesIndex}_${data_processed.length - 1 - itemPos}`).style('stroke-width', 1);
    };

    chart.defaultMouseOutFunc = function (d, i) {
        if (!gElementsCollected) {
            collectGroups(commonRendererProp.params.rendererIndex);
        }

        if (isPatternPalleteEnabled) {
            funnelsections.style('opacity', null);
            ellipseSections.style('opacity', null);
        } else {
            funnelsections.style('fill', function (d, i) {
                var data = defined(d.data) ? d.data : d,
                isSelection = defined(chartObj.cache.clickSelectionData);

                if(isSelection && (isUndefined(data.clickSelection) || !data.clickSelection) && defined(chartObj.cache.clickSelectionData[0])){//To maintain selection effect{
                    return chartObj.systemConf.chart.plot.eventHandler.click.behaviour.unselected;
                    ;
                }

                return (isFacet(chartObj) ? fillColor[i] : fillColor[0])(i, d);
            });
            ellipseSections.style('fill', function (d, i) {
                var data = defined(d.data) ? d.data : d,
                isSelection = defined(chartObj.cache.clickSelectionData);

                if(isSelection && (isUndefined(data.clickSelection) || !data.clickSelection) && defined(chartObj.cache.clickSelectionData[0])){//To maintain selection effect{
                    return chartObj.systemConf.chart.plot.eventHandler.click.behaviour.unselected;
                }

                return (isFacet(chartObj) ? fillColor[i] : fillColor[0])(i, d, 0.5);
            });
        }
        label.selectAll('text').style('text-shadow', null);
        labelLinkgroup.selectAll(PATH_ELEMENT).style('stroke-width', null);
    };

    function tweenAll(d, i, elementType, animationType) {
        var base = baseData(animationType, i),
            dt = trapezoid[i];
        var interpolate =
            elementType == 'funnel'
                ? d3_interpolate(drawTrapezoid(base), drawTrapezoid(dt))
                : d3_interpolate(drawEllipse(base), drawEllipse(dt));
        return function (t) {
            return interpolate(t);
        };
    }

    function baseData(animationType, i) {
        var d = simpleClone(trapezoid[animationType == VERTICALALL_ANIMATION ? 0 : i]);
        if (animationType === HORIZONTALALL_ANIMATION) {
            d.topLeft[0] = d.middleLeft[0] = d.bottomLeft[0] = d.topRight[0] = d.middleRight[0] = d.bottomRight[0] = cfg.width / 2;
        } else {
            d.middleLeft = d.bottomLeft = d.topLeft;
            d.middleRight = d.bottomRight = d.topRight;
        }
        return d;
    }

    function hideDisabledElements(data, dur) {
        ellipseSections.each(function (d, i) {
            var hide = data[i].disabled || !dataset.getY(data_processed[i], yaxiscolumnorder);
            var t = hide ? dur : 0;
            if (t) {
                setTimeout(function () {
                    hideSection(i, hide);
                }, t);
            } else {
                hideSection(i, hide);
            }
        });
    }

    function hideSection(i, hide) {
        var display = hide ? NONE : null;

        ellipse.selectAll(`#ellipse${i}`).style('display', display);
        funnel.selectAll(`#funnel${i}`).style('display', display);
    }

    function preCalArcHeight() {
        var h = chartHeight - 2 * outerPadding;
        var arcHeight = h * 0.05;
        var padding = getFunnelPadding(h);
        arcHeight = (arcHeight - data_processed.length) / data_processed.length;
        arcHeight = mathMax(minArcHeight, arcHeight);
        arcHeight = !threeDimension ? 0 : cfg.canOverlap ? arcHeight + maxPadding / 2 : padding / 2 - arcCorrection;
        return arcHeight;
    }

    function getFunnelPadding(h) {
        var userPadding = commonPlotOptions.padding;
        if (defined(userPadding)) {
            return invokeFunction(userPadding, h, noOfGaps, chartObj);
        }

        var maxPaddingHeight = h * percentage;
        return mathMin(maxPadding, mathMax(minPadding, maxPaddingHeight / noOfGaps));
    }

    /**
     * This function is used to find the point of the arc to get border radius.
     * @param {Array[2]} point1 - [x1, y1] - Starting point of the line
     * @param {Array[2]} point2 - [x2, y2] - Ending point of the line
     * @param {Number} borderRadius - Expected border radius of the starting point
     * @returns {Array[2]} Coordinate of the arc near the starting point
     */
    function findPointOnLine(point1, point2, borderRadius) {
        const [x1, y1] = point1, [x2, y2] = point2;
        const L = Math.sqrt((x2 - x1) ** 2 + (y2 - y1) ** 2)
        const mx = (x2 - x1) / L;   // x-slope
        const my = (y2 - y1) / L;   // y-slope
        const dx = borderRadius * mx;   // change in x position due to arc
        const dy = borderRadius * my;   // change in y position due to arc
        const x = x1 + dx;  // Arc intersection points [x, y]
        const y = y1 + dy; //TODO -dx & -dy
        return [x, y];
    }

    function drawTrapezoid(d) {
        var topW = (d.topRight[0] - d.topLeft[0]) / 2,
            bottomW = (d.bottomRight[0] - d.bottomLeft[0]) / 2;
        var topCurve = cfg.canOverlap ? ((2 * topW) / cfg.width) * arcHeight : arcHeight1,
            bottomCurve = cfg.canOverlap ? ((2 * bottomW) / cfg.width) * arcHeight : arcHeight1;

        if (!threeDimension && (Object.keys(d.border).length > 0 || (d.index === 0 && d.stemRadius !== null))) {
            const dTopLeft = d.topLeft, dTopRight = d.topRight, dBottomLeft = d.bottomLeft, dBottomRight = d.bottomRight;
            const isShowFunnel = mathAbs(dTopRight[0] - dBottomRight[0]) > 0.01,
                stemRadius = d.stemRadius,
                borderTopLeft = (isShowFunnel) ? (d.index === 0 && stemRadius !== null) ? stemRadius : (d.border.topLeft || 0) : 0,
                borderTopRight = (isShowFunnel) ? (d.index === 0 && stemRadius !== null) ? stemRadius : (d.border.topRight || 0) : 0,
                borderBottomLeft = (isShowFunnel && d.border.bottomLeft) ? d.border.bottomLeft : 0,
                borderBottomRight = (isShowFunnel && d.border.bottomRight) ? d.border.bottomRight : 0,
                isSmallTail = ( (stemRadius ? stemRadius : borderTopLeft ? borderTopLeft : borderTopRight ? borderTopRight : 0) - mathAbs(dTopRight[0] - dTopLeft[0])) >= 0,
                pointTopLeft = (isShowFunnel) ? findPointOnLine(dTopLeft, dBottomLeft, borderTopLeft) : dTopLeft,
                pointTopRight = (isShowFunnel) ? findPointOnLine(dTopRight, dBottomRight, borderTopRight) : dTopRight,
                pointBottomLeft = (isShowFunnel) ? findPointOnLine(dBottomLeft, dTopLeft, borderBottomLeft) : dBottomLeft,
                pointBottomRight = (isShowFunnel) ? findPointOnLine(dBottomRight, dTopRight, borderBottomRight) : dBottomRight;

            return `M ${(isSmallTail) ? `${pointTopLeft[0]} ${pointTopLeft[1]}` : `${dTopLeft[0] + borderTopLeft} ${dTopLeft[1]}`}
                    A${topW} ${topCurve} 0 0 0  ${(isSmallTail) ? `${pointTopLeft[0]} ${pointTopLeft[1]}` : `${dTopRight[0] - borderTopRight} ${dTopRight[1]}`}
                    A${borderTopRight / 2 * ((isSmallTail) ? 1 : 3)} ${borderTopRight / 2 * ((isSmallTail) ? 1 : 3)} 0 0 0 ${pointTopRight[0]} ${pointTopRight[1]}
                    L${pointBottomRight[0]} ${pointBottomRight[1]} 
                    A${borderBottomRight / 3 * 2} ${borderBottomRight / 3 * 2} 0 0 0 ${dBottomRight[0] - borderBottomRight} ${dBottomRight[1]}
                    A${bottomW} ${bottomCurve} 0 0 1 ${dBottomLeft[0] + borderBottomLeft} ${dBottomLeft[1]} 
                    A${borderBottomLeft / 3 * 2} ${borderBottomLeft / 3 * 2} 0 0 0 ${pointBottomLeft[0]} ${pointBottomLeft[1]}
                    L${pointTopLeft[0]} ${pointTopLeft[1]}
                    A${borderTopLeft * 3 / 2} ${borderTopLeft * 3 / 2} 0 0 0 ${(isSmallTail) ? `${pointTopLeft[0]} ${pointTopLeft[1]}` : `${dTopLeft[0] + borderTopLeft} ${dTopLeft[1]}`}
                    L ${(isSmallTail) ? `${pointTopLeft[0]} ${pointTopLeft[1]}` : `${dTopRight[0] - borderTopRight} ${dTopRight[1]}`}
                    Z`;
            // }
            // return `M${d.topLeft[0] + borderTopLeft} ${d.topLeft[1]}
            //         A${topW} ${topCurve} 0 0 0 ${d.topRight[0] - borderTopRight} ${d.topRight[1]} 
            //         A${borderTopRight * 3 / 2} ${borderTopRight * 3 / 2} 0 0 0 ${pointTopRight[0]} ${pointTopRight[1]}
            //         L${pointBottomRight[0]} ${pointBottomRight[1]} 
            //         A${borderBottomRight / 3 * 2} ${borderBottomRight / 3 * 2} 0 0 0 ${d.bottomRight[0] - borderBottomRight} ${d.bottomRight[1]}
            //         A${bottomW} ${bottomCurve} 0 0 1 ${d.bottomLeft[0] + borderBottomLeft} ${d.bottomLeft[1]} 
            //         A${borderBottomLeft / 3 * 2} ${borderBottomLeft / 3 * 2} 0 0 0 ${pointBottomLeft[0]} ${pointBottomLeft[1]}
            //         L${pointTopLeft[0]} ${pointTopLeft[1]}
            //         A${borderTopLeft * 3 / 2} ${borderTopLeft * 3 / 2} 0 0 0 ${d.topLeft[0] + borderTopLeft} ${d.topLeft[1]}
            //         L${d.middleLeft[0]} ${d.middleLeft[1]}
            //         Z`;
        }

        return `M${d.topLeft[0]} ${d.topLeft[1]} 
                A${topW} ${topCurve} 0 0 0 ${d.topRight[0]} ${d.topRight[1]} 
                L${d.middleRight[0]} ${d.middleRight[1]} 
                L${d.bottomRight[0]} ${d.bottomRight[1]} 
                A${bottomW} ${bottomCurve} 0 0 1 ${d.bottomLeft[0]} ${d.bottomLeft[1]} 
                L${d.middleLeft[0]} ${d.middleLeft[1]}
                Z`;
    }

    function drawEllipse(d) {
        var bottomW = (d.bottomRight[0] - d.bottomLeft[0]) / 2;
        var bottomCurve = cfg.canOverlap ? ((2 * bottomW) / cfg.width) * arcHeight + arcCorrection : arcHeight1;
        // const isShowFunnel = Math.abs(d.topRight[0] - d.bottomRight[0]) > 0.01,
        //     borderBottomLeft = (isShowFunnel && d.border.bottomLeft) ? d.border.bottomLeft : 0,
        //     borderBottomRight = (isShowFunnel && d.border.bottomRight) ? d.border.bottomRight : 0;

        // return `M${d.bottomLeft[0] + borderBottomLeft} ${d.bottomLeft[1]} A${bottomW} ${bottomCurve} 0 0 1 ${d.bottomRight[0] - borderBottomRight} ${d.bottomRight[1]
        //     } A${bottomW} ${cfg.canOverlap ? bottomCurve : bottomCurve + arcCorrection} 0 0 1 ${d.bottomLeft[0] + borderBottomLeft} ${d.bottomLeft[1]}`;
        return `M${d.bottomLeft[0]} ${d.bottomLeft[1]} A${bottomW} ${bottomCurve} 0 0 1 ${d.bottomRight[0]} ${d.bottomRight[1]}
                A${bottomW} ${cfg.canOverlap ? bottomCurve : bottomCurve + arcCorrection} 0 0 1 ${d.bottomLeft[0]} ${d.bottomLeft[1]}`;
    }

    function collectGroups(rendererIndex = 0) {
        let subSeriesIndex = chart.subSeriesIndex || 0,
            parentWrap = isFacet(chartObj) ? chartObj.staticPaneSVG.selectAll(`g.grid${paneid}`) : chartObj.staticPaneSVG;

        labelLinkgroup = labelLinkgroup || parentWrap.selectAll(`g.link_${rendererIndex} g.subseries_${subSeriesIndex}`);
        label = label || parentWrap.selectAll(`g.label_${rendererIndex} g.subseries_${subSeriesIndex}`);
        gElementsCollected = true;
    }

    chart.shiftInfo = function(_) {
        return arguments.length ? (shiftInfo = _, chart): shiftInfo;
    }

    return chart;
};

export var _funnelChart = svg_plot_renderer.funnel;
