// @zohocharts/zc/src/legend/shape-legend/list/shape-list.js
//
import { Registry } from '../../../core/zc-registry';
import { LegendRegistry } from '../../legend-registry';
import { BaseLegend } from '../../base-legend';
import { ListLegendParser } from '../../list-legend/list-legend-parser';
import { NULL } from '../../../core/zc-constants';
import { getMaxToResize } from '../../legend-bound';
import { getVisibleItemRenderCount } from '../../size-legend/list/expandable-helper';

class ShapeLegendList extends BaseLegend {
    create() {
        let viewScaleComponent = this.viewScaleComponent();
        this.spec.data = viewScaleComponent.ticks;
        let _expand_info = getVisibleItemRenderCount(this);
        this.spec.visibleItemCount = _expand_info && _expand_info.visibleItemCount;
        this.spec.scales = { shape: viewScaleComponent.scale };

        let mark = ListLegendParser(this.spec, viewScaleComponent, this); //TODO: remove 'this'
        this._parser = mark._helper;
        return (this.encode = mark);
    }
    draw(parent) {
        parent.styles({
            display: 'block',
            position: 'absolute',
            left: NULL, //Reset left and top position. so that floating legend won't get trimmedon redraw (legend-03.json)
            top: NULL
        });
        super.draw(parent);
    }

    resize(width, height) {
        if (!width && !height) {
            return;
        }

        let viewScaleComponent = this.viewScaleComponent();
        let _expand_info = getVisibleItemRenderCount(this);
        this.spec.visibleItemCount = _expand_info && _expand_info.visibleItemCount;

        let { w, h } = getMaxToResize(width, height, this.spec.bound);
        this.updateScale(w, h);
        super.resize(w, h);
    }
}

ShapeLegendList.encode = 'shape';
ShapeLegendList.view = 'list';
ShapeLegendList.domMode = 'html';

export { ShapeLegendList };
Registry.setComponent('shapeLegendList', ShapeLegendList);
LegendRegistry.add(ShapeLegendList);
