// @zohocharts/zc/src/mark/dom-paint.js
//
import { isSvgElement } from '../utils/zcsvgutils';
import { pick, each, isObject, getVal, arrayIncludes } from '../utils/zcutils';
import { cssProperties } from './helper/css-property-names';
import { getDashArray, getLineCap } from '../utils/dashstyles';
import { config_CSS_value_converter, parseMarkConfig } from './helper/mark-config-parser';
import { customMarkMap } from './helper/custom-mark/custom-mark-map';
import { getExpressionValue, isExpressionType } from './../mark/parser/parser-helpers';
import { ALLOWED_REPARSE_MARKS_CONFIG, ALLOWED_REPARSE_OPERATORS } from './operators/config/allowed-config-operators';
import { Registry } from '../core/zc-registry';

function paintDom(properties, ele) {
    if (ele.empty()) {
        return;
    }
    const isHTMLType = (this?.type || '').toLowerCase() == 'html';

    each(properties, (value, prop) => {
        if (prop === 'styles') {
            value = propertiesToStyles(value);
            if (isSvgElement(ele) && !isHTMLType) {
                value = enhanceStyles(value);
            }
        } else if (prop === 'props') {
            return;
        }

        // if (isObject(value)) {
        //     var a = prop.substr(0, prop.length - 1);
        //     for (var attr in value) {
        //         ele[a](attr, value[attr]);
        //     }
        // } else {
        //     ele[prop](value);
        // }
        ele[prop](value);
    });
}

function paintDomByConfig(obj, ele, cache, morph, parent) {
    let { mark, config } = obj,
        output = {};

    if (ele.empty() || !config) {
        return;
    }

    config = parseMarkConfig(mark, config, cache, morph, parent);
    let keys = customMarkMap[mark] ? Object.keys(customMarkMap[mark](config).depends) : [];

    each(config, (value, prop) => {
        let { attrType = 'attr', cssName = prop, value: valueFunction } = cssProperties[prop] || {},
            _value = valueFunction
                ? function (d, i) {
                      return valueFunction(getVal(value, [d, i], this));
                  }
                : value;
               let isExpression = isExpressionType(_value);
               if(isExpression){
                    _value = function(d, i){
                        let __viewflow =  cache ? cache.__tokens__ || cache.cache.__tokens__ : null;
                        let v = getExpressionValue(this,[d, i] , value, prop, __viewflow);
                        if(config_CSS_value_converter[prop]){
                            v = config_CSS_value_converter[prop](config, v)
                        }
                        return v;
                    }
               }
        if (attrType === 'attr' || attrType === 'style') {
            if (isObject(cssName)) {
                cssName = isSvgElement(ele) ? cssName.svg : cssName.html;
            }
            output[`${attrType}s`] = output[`${attrType}s`] || {};
            output[`${attrType}s`][cssName] = _value;
        } else {
            output[attrType] = _value;
        }

        ele.each(function (d, i) {
            let node = this;
            if (node.zc_cache) {
                if (!arrayIncludes(keys, prop)) {
                    node.zc_cache[prop] = _value;
                }
            }
        });
    });

    for (let attr in output) {
        ele[attr](output[attr]);
    }

    //Reparse the dom based on the config - if any (as of now only text-overflow is handled)
    rePaintDomByConfig(mark, config, ele, cache); 
    
}

function rePaintDomByConfig(mark, config, nodes, cache) {
    let reparse_config = ALLOWED_REPARSE_MARKS_CONFIG.get(mark);
    if(reparse_config){
        reparse_config.map((d) => {
            if(config[d]){
                let componentName = ALLOWED_REPARSE_OPERATORS.get(d),
                    operator = Registry.getComponent(componentName);
                if(operator){
                    operator.call(null, nodes, d, config, cache);
                }
            }
        })
    }
}
function propertiesToStyles(properties) {
    var styles = {};
    each(properties, (value, prop) => (styles[svgStyles[prop] || prop] = value));
    return styles;
}

function enhanceStyles(styles) {
    styles['stroke-width'] = pick(styles['stroke-width'], 1);
    styles['stroke-linecap'] = getLineCap(styles['stroke-linecap'], styles['stroke-dasharray']);
    styles['stroke-dasharray'] = getDashArray(styles['stroke-dasharray'], styles['stroke-width']);

    return styles;
}

let svgStyles = {
    //Fill styles
    fillColor: 'fill',
    fillOpacity: 'fill-opacity',
    //Stroke styles
    strokeColor: 'stroke',
    strokeOpacity: 'stroke-opacity',
    strokeWidth: 'stroke-width',
    lineCap: 'stroke-linecap',
    dashStyle: 'stroke-dasharray'
};

export function reStyleDom(node, handler, cache){
    paintDomByConfig({config: handler.config}, node,{cache});
    paintDom(handler.properties, node);
}
export { paintDom, paintDomByConfig, propertiesToStyles };
