// @zohocharts/zc/src/mark/dom-animation.js
//
import { Registry } from '../core/zc-registry';
import { d3_ease, d3_interpolate, d3_select } from '../core/d3-constants';
import { splat, arrayIncludes, each, getValByPriority, pick, getVal, invokeFunction, isArray, defined } from '../utils/zcutils';
import { isFunction } from '../utils/function-utils';
import { paintDomByConfig } from './dom-paint';
import { parseMarkConfig } from './helper/mark-config-parser';
import { cssProperties } from './helper/css-property-names';
import { _defaultMarkValue } from './parser/parser-helpers';

function animeDom(config, parser, element, cache, mark, final, rootParser) {
    if (element.empty() || !config) {
        return;
    }

    final = final || (config.on && config.on.end) || config.onEnd;
    config = splat(config);
    let args = { arr: config, i: 0, element, cache, mark, parser, rootParser };
    d3ChainAnimation(args.arr, args, final);

    // let terminator = (_, i) => {
    //         if (_.i >= config.length || i === config.length - 1) {
    //             final && final();
    //             // invokeFunction(parser.onEnd, element, parser);
    //             return true;
    //         }
    //     },
    //     onEnd = (_) => _.i++;
    // chainPromise({ args, terminator, onEnd, callback: simpleAnime });
}

function d3ChainAnimation(config, args, onEnd) {
    let { element, cache, mark, parser, rootParser } = args,
        transition;

    let transition_ref = args.cache.__tokens__ ? args.cache.__tokens__.transition : null;
    
    config.forEach((_, i) => {
        if(transition_ref){
            transition_ref.count++;
            transition_ref.start = true;
        }
        
        if (isFunction(_)) {
            if (transition) {
                transition.on('end', () => {
                    _(element, transition)
                    if(transition_ref){
                        transition_ref.count--;
                        if(transition_ref.count == 0){
                            transition_ref.end = true;
                        }
                    }
                    
                });
            } else {
                _(element, transition);
            }
        } else if (isArray(_)) {
            if (transition) {
                transition.on('end', () => 
                    {
                        syncAnimation(_, args);
                        if(transition_ref){
                            transition_ref.count--;
                            if(transition_ref.count == 0){
                                transition_ref.end = true;
                            }
                        }
                        
                    });
            } else {
                syncAnimation(_, args);
            }
        } else {
            let { from, to, interpolate, property } = getConfigStops(_),
                { duration, delay, easing, transition: transitionName } = d3AnimeConfig(_, parser);
            let _cache = {cache, parser: rootParser}
            applyConfig(transition || element, from, _cache, mark, rootParser);
            applyProperties(transition || element, property.from);

            transition = (transition || element).transition(transitionName);
            if (!defined(transitionName)) {
                transition
                    .duration(duration)
                    .delay(delay)
                    .ease(d3_ease(easing));
            }
            applyConfig(transition, to, _cache, mark, rootParser);
            applyProperties(transition, property.to);

            if (!isEmptyObject(interpolate)) {
                tween(transition, interpolate, cache, mark);
            }

            if (i === config.length - 1) {
                transition.on('end', ()=>{
                    if(onEnd){
                        onEnd();
                    }

                    if(transition_ref){
                        transition_ref.count--;
                        if(transition_ref.count == 0){
                            transition_ref.end = true;
                        }
                    }
                    
                    
                });
            }else{
                transition.on('end', ()=>{
                    if(transition_ref){
                        transition_ref.count--;
                        if(transition_ref.count == 0){
                            transition_ref.end = true;
                        }
                    }
                    
                })
            }
        }
    });
}

function syncAnimation(config, args) {
    let { element, cache, mark, parser } = args,
        node = element.node();

    if (element.empty()) {
        return;
    }

    config.forEach(_ => {
        let { from, to, interpolate } = getConfigStops(_),
            { duration, delay, easing, transition: transitionName } = d3AnimeConfig(_, parser),
            a1 = {};

        each(to, (value, prop) => {
            a1[prop] = pick(from[prop], node.zc_cache[prop], _defaultMarkValue[prop]);
        });

        let _from = parseMarkConfig(mark, a1, cache, true),
            _to = parseMarkConfig(mark, to, cache, true);

        let transition = d3_select({}).transition(transitionName);
        if (!defined(transitionName)) {
            transition
                .duration(duration)
                .delay(delay)
                .ease(d3_ease(easing));
        }
        transition.tween('style', function(d, i) {
            let _i = {};

            each(_to, (value, prop) => {
                let { attrType = 'attr', cssName = prop, value: valueFunction } = cssProperties[prop] || {}, // TODO: given by user
                    _value = _ => (valueFunction ? valueFunction(getVal(_, [d, i], node)) : getVal(_, [d, i], node));
                _i[prop] = d3_interpolate(_value(_from[prop]), _value(value));
            });
            return function(t) {
                each(_to, (value, prop) => {
                    let { attrType = 'attr', cssName = prop, value: valueFunction } = cssProperties[prop] || {};

                    if (attrType === 'attr' || attrType === 'style') {
                        element[attrType](cssName, _i[prop](t));
                    } else {
                        element[attrType](_i(t));
                    }
                });
            };
        });
    });
}

function getConfigStops(_) {
    let _from = {},
        _to = {},
        _interpolate = {},
        _property = { from: {}, to: {} },
        _default = ['duration', 'delay', 'easing', 'transition'];

    for (let attr in _) {
        if (_.hasOwnProperty(attr)) {
            let cur = _[attr];
            if (!arrayIncludes(_default, attr) && defined(cur)) {
                let { property, from, to, interpolator } = cur;
                if (property && cur.hasOwnProperty('property')) {
                    from !== undefined && (_property.from[attr] = [property, from]);
                    _property.to[attr] = [property, to];
                } else {
                    if (!interpolator) {
                        from !== undefined && (_from[attr] = from);
                        _to[attr] = to !== undefined ? to : cur;
                    }
                    _interpolate[attr] = cur;
                }
            }
        }
    }
    return { from: _from, to: _to, interpolate: _interpolate, property: _property };
}

function applyConfig(element, config, cache, mark, parser) {
    if (!config || isEmptyObject(config)) {
        return;
    }
    paintDomByConfig({ mark, config }, element, cache, true);
}

function applyProperties(element, property) {
    if (!property || isEmptyObject(property)) {
        return;
    }
    for (var attr in property) {
        let cur = property[attr];
        element[cur[0]](attr, cur[1]);
    }
}

function isEmptyObject(obj) {
    for (let name in obj) {
        return false;
    }
    return true;
}

function tween(ele, config, cache, mark) {
    if (ele.empty() || !config) {
        return;
    }

    let from = {},
        to = {},
        interpolator = {};
    for (let attr in config) {
        if (config[attr].from !== undefined) {
            from[attr] = config[attr].from;
        }
        if (config[attr].to !== undefined) {
            to[attr] = config[attr].to;
        } else {
            to[attr] = config[attr];
        }
        if (config[attr].interpolator !== undefined) {
            interpolator[attr] = config[attr].interpolator;
        }
    }

    if (isEmptyObject(interpolator)) {
        return;
    }

    let _from = parseMarkConfig(mark, from, cache, true),
        _to = parseMarkConfig(mark, to, cache, true);

    each(_to, (value, prop) => {
        let { attrType = 'attr', cssName = prop, value: valueFunction } = cssProperties[prop] || {};
        // _value = valueFunction
        //     ? function(d, i) {
        //           return valueFunction(getVal(value, [d, i], this));
        //       }
        //     : value;
        //
        ele[`${attrType}Tween`](prop, function tween(d, i, j) {
            let a1 = {};
            let node = this;
            each(to, (value, prop) => {
                let _value = getVal(value, [d, i], this);
                a1[prop] = (interpolator[prop] || d3_interpolate)(pick(from[prop], node.zc_cache[prop], _defaultMarkValue[prop]), _value);
            });

            return function(t) {
                let a = {};
                each(a1, (value, prop) => {
                    a[prop] = a1[prop](t);
                });
                return getVal(parseMarkConfig(mark, a, cache, true)[prop], [d, i], node);
            };
        });
    });
}

// function chainPromise(_) {
//     let { args, terminator, onEnd, callback } = _;
//     // callback(args);
//     // return;
//
//     if (terminator && terminator(args)) {
//         return;
//     }
//
//     let _promise = new Promise((resolve, reject) => {
//         callback(args, () => {
//             onEnd(args);
//             resolve(_);
//         });
//     }).then(chainPromise);
// }

// function simpleAnime(args, onEnd) {
//     let { arr, i, element, cache, mark, parser } = args,
//         config = arr[i],
//         { duration, delay, easing, transition: transitionName } = d3AnimeConfig(config, parser),
//         transition = element.transition(transitionName);
//
//     if (isFunction(config)) {
//         config(element, transition);
//         onEnd(); // Check for args
//         return;
//     }
//
//     if (isArray(config)) {
//         return syncAnimation(config, args);
//     }
//
//     let { from, to, interpolate } = getConfigStops(config);
//
//     applyConfig(element, from, cache, mark);
//
//     transition
//         .duration(duration)
//         .delay(delay)
//         .ease(d3_ease(easing))
//         .on('end', function (d, i) {
//             let _this = this;
//             setTimeout(function () {
//                 onEnd.call(_this, d, i);
//             }, 0);
//         });
//
//     applyConfig(transition, to, cache, mark);
//
//     if (!isEmptyObject(interpolate)) {
//         tween(transition, interpolate, cache, mark);
//     }
// }

function d3AnimeConfig(config, globalConfig) {
    return {
        duration: pick(getValByPriority([config, globalConfig], 'duration'), 250),
        delay: getValByPriority([config, globalConfig], 'delay') || 0,
        easing: getValByPriority([config, globalConfig], 'easing'),
        transition: getValByPriority([config, globalConfig], 'transition')
    };
}

export { animeDom };
Registry.setComponent('mark.anime', animeDom);
