// @zohocharts/zc/src/legend/size-legend/slider/size-slider-options.js
import { CENTER_ALIGN } from '../../../core/align-constants';
import { Registry } from '../../../core/zc-registry';

export let sizeSliderOptions = userProp => {
    return {
        marker: {
            shape: 'circle',
            hAlign: CENTER_ALIGN,
            vAlign: CENTER_ALIGN,
            fillColor: '#51545A',
            strokeColor: 'white'
        },
        track: {
            show: true,
            fillColor: '#F2F3F5'
        },
        connector: {
            show: true,
            strokeColor: '#51545A'
        },
        axis: {
            show: true,
            ticklabel: {
                fontSize: userProp.fontL1Size.l5
            },
            minmaxlabels: {
                padding: 2,
                align: 'center',
                fontSize: userProp.fontL1Size.l5
            },
            tickmark: {
                color: 'rgb(97, 97, 97)',
                size: 5
            },
            axisline: {
                show: true,
                color: 'rgb(97, 97, 97)'
            }
        }
    };
};

Registry.setComponent('sizeSliderOptions', sizeSliderOptions);
