// @zohocharts/zc/src/legend/size-legend/slider/size-slider-parser.js
//
import { HORIZONTAL_ALIGN, BOTTOM_ALIGN, RIGHT_ALIGN, LEFT_ALIGN } from '../../../core/align-constants';
import { TRACK_LEGEND_CLASSNAME, SLIDER_VIEW_CLASSNAME } from '../../legend-constants';
import { TRANSPARENT } from '../../../core/zc-constants';
import { d3_selectAll, d3_max, d3_merge } from '../../../core/d3-constants';
import { Mark } from '../../../mark/mark-utils';
import { LegendTitleSvgParser } from '../../title/legend-title-svg-parser';
import { _LegendBackgroundSvgParser } from '../../background/legend-background-svg-parser';
import { LegendContentWrapParser } from '../legend-content-wrap-parser';
import { TrackerParser } from './tracker-parser';
import { ConnectorParser } from './slider-connector-parser';
import { SliderAxisParser } from './slider-axis-parser';
import { MinmaxLabelParser } from '../../minmax-label/legend-minmax-label-parser';
import { createElement } from '../../../utils/zcsvgutils';
import { pick, parseShortHandValue, bbox, d3_values, getDistinct, simpleClone } from '../../../utils/zcutils';
import { getSVGsymbol } from '../../../utils/svg-symbols';
import { getSliderSize } from '../../legend-minmaxlabels';
import { getLegendAlignment } from '../../legend-utils';
import { getDisabledOptions } from '../../events/legend-event-helpers';
import { handleTicklabels } from '../nested/size-nested-ticklabel-handler';

function SizeSliderParser(obj) {
    let _helper = helper(obj);
    let { titleSpec, backgroundSpec, trackSpec, axisSpec, minmaxLabelSpec, connectorSpec, selectionTrackSpec } = _helper;

    // Wrap
    let groupMark = Mark('group', '.legendgroup'),
        contentMark = LegendContentWrapParser(),
        //
        // Title
        titleMark = LegendTitleSvgParser(titleSpec),
        // Tracker
        trackerMark = TrackerParser(trackSpec),
        // Selection tracker
        selectionTrackerMark = TrackerParser(selectionTrackSpec),
        // Axis
        sliderAxisMark = SliderAxisParser(axisSpec),
        // Connector
        connectorMark = ConnectorParser(connectorSpec),
        // Slider
        sliderViewMark = Mark('group', `.${SLIDER_VIEW_CLASSNAME}`),
        // Minmax labels
        minmaxLabelMark = MinmaxLabelParser(minmaxLabelSpec),
        // Background
        backgroundMark = _LegendBackgroundSvgParser(backgroundSpec);

    contentMark.onEnd = obj.align;
    contentMark.children = [titleMark, trackerMark, selectionTrackerMark, sliderAxisMark, connectorMark, sliderViewMark, minmaxLabelMark];
    groupMark.children = [backgroundMark, contentMark];
    groupMark._helper = _helper;

    return groupMark;
}

function helper(obj) {
    let { spec, viewScaleComponent, brushOptions } = obj;
    let { config, bound, format, globalFont } = spec,
        // View scale component
        { scale: viewScale, bound: viewBound, ticks, labels } = viewScaleComponent,
        data = labels || ticks,
        //
        { slider: sliderConfig, stops, ranges } = config,
        { axis: axisConfig, marker: markerConfig, track: trackConfig } = sliderConfig,
        filterConfig = getDisabledOptions(config),
        padding = pick(sliderConfig.padding, 3),
        axisPadding = axisConfig.show ? pick(axisConfig.padding, 3) : 0,
        parsedPadding = parseShortHandValue(padding),
        //
        align = getLegendAlignment(config),
        isHorizontal = align.layout === HORIZONTAL_ALIGN,
        axisOrient = axisConfig.orient || (isHorizontal ? BOTTOM_ALIGN : RIGHT_ALIGN),
        shape = markerConfig.shape;

    let { reversed, brushScale, dimension, markerBound, singleDomainContinuous, sizeScale } = brushOptions,
        dim = { width: dimension[0], height: dimension[1] },
        _extentData = [data[0], data[data.length - 1]],
        extentData = reversed ? d3_values(_extentData).reverse() : _extentData,
        minmaxLabelData = isHorizontal ? extentData : d3_values(extentData).reverse(), // Minmax label expects like this,
        axisData = singleDomainContinuous
            ? sizeScale.domain()
            : stops && !ranges
            ? getDistinct(d3_merge([stops, brushScale.domain()]))
            : brushScale.domain(),
        sliderSize = minmaxLabelData.map(d => [viewScale(d), viewScale(d)]),
        markerSize,
        minmaxLabelDim = dimension,
        availableBound = {
            maxWidth: bound.maxWidth - parsedPadding.left - parsedPadding.right,
            maxHeight: bound.maxHeight - parsedPadding.top - parsedPadding.bottom
        };

    if (filterConfig.type) {
        let temp = getSliderSize(filterConfig, dimension, null, align),
            _pos = isHorizontal ? 1 : 0;
        sliderSize = sliderSize.map(_ => [temp[0], _[_pos]]);
    } else if (brushOptions.handles && shape !== 'circle' && shape !== 'square') {
        // Minmax labels need exact slider size
        let tempEle = createElement('path', '.temp_slider_marker', [1], d3_selectAll('svg.svgPattern'));
        let symbolPath = getSVGsymbol(shape),
            _b = [];
        sliderSize = minmaxLabelData.map(d => {
            tempEle.attr('d', symbolPath(markerBound.width(d), markerBound.height(d)));
            let b = bbox(tempEle);
            _b.push(b);
            return isHorizontal ? [b.width, b.width] : [b.height, b.height];
        });
        tempEle.remove();
        minmaxLabelDim = isHorizontal ? [dimension[0], d3_max(_b, _ => _.height)] : [d3_max(_b, _ => _.width), dimension[1]];
    }

    markerSize = isHorizontal ? simpleClone(sliderSize) : simpleClone(sliderSize).reverse();
    let availableAxisBound = getAxisBound(availableBound, minmaxLabelDim, isHorizontal, axisOrient, axisPadding);

    let bandWidth = brushScale.bandwidth ? brushScale.bandwidth() : 0;
    if (bandWidth) {
        // ZC2496
        sliderSize.forEach(_ => {
            _[0] -= bandWidth;
            _[1] -= bandWidth;
        });
    }

    let titleSpec = { data: [1], config: config.title, bound },
        backgroundSpec = { data: [1], config: config, margin: padding },
        alignSpec = { isHorizontal, padding, axisPadding, axisOrient, markerSize },
        trackSpec = {
            data: extentData,
            scale: viewScale,
            brushScale,
            config: sliderConfig,
            filterConfig,
            trackConfig: { fillColor: TRANSPARENT, ...filterConfig.track, show: trackConfig.show },
            margin: padding,
            bound: dim,
            markerBound,
            align
        },
        selectionTrackSpec = {
            ...trackSpec,
            trackConfig,
            className: `${TRACK_LEGEND_CLASSNAME}-selection`
        },
        axisSpec = {
            data: axisData,
            format,
            scale: brushScale,
            config: axisConfig,
            bound: dim,
            viewBound: isHorizontal
                ? { width: viewBound[0], height: minmaxLabelDim[1] }
                : { width: minmaxLabelDim[0], height: viewBound[1] },
            align,
            orient: axisOrient,
            availableBound: availableAxisBound
        },
        minmaxLabelSpec = {
            format, //TODO
            data: minmaxLabelData,
            index: isHorizontal ? undefined : i => (i ? 0 : 1),
            globalFont,
            config: axisConfig.minmaxlabels,
            orient: axisOrient,
            dimension: minmaxLabelDim,
            sliderSize: config.filter.enabled ? sliderSize : [0, 0],
            trimText: ele => {
                let labelEle = ele.selectAll('text'),
                    minmaxLabelAlign = axisConfig.minmaxlabels.align,
                    _b = getMinmaxLabelBound(
                        availableBound,
                        minmaxLabelDim,
                        minmaxLabelAlign,
                        isHorizontal,
                        axisOrient,
                        dimension,
                        bound,
                        sliderSize
                    ),
                    textAnchor = minmaxLabelAlign === RIGHT_ALIGN ? 'start' : minmaxLabelAlign === LEFT_ALIGN ? 'end' : 'middle',
                    _labelParser = { anchor: textAnchor };

                handleTicklabels(ele, ele.selectAll('.nothing'), _b, { width: 0, height: 0 }, _labelParser);
            }
        },
        connectorSpec = {
            config: sliderConfig,
            data: extentData,
            align,
            bound: dim,
            scale: viewScale,
            brushScale,
            filterConfig,
            markerBound
        };

    return {
        titleSpec,
        backgroundSpec,
        trackSpec,
        axisSpec,
        minmaxLabelSpec,
        connectorSpec,
        alignSpec,
        selectionTrackSpec
    };
}

function getMinmaxLabelBound(available, minmaxLabelDim, align, isHorizontal, orient, dimension, bound, sliderSize) {
    let { maxWidth, maxHeight } = available,
        x;

    if (isHorizontal) {
        return { maxWidth, maxHeight, x: (bound.maxWidth - dimension[0] - sliderSize[0][0] / 2 - sliderSize[1][0] / 2) / 2 };
    }

    if (orient === LEFT_ALIGN && align === RIGHT_ALIGN) {
        maxWidth -= minmaxLabelDim[0];
        x = minmaxLabelDim[0];
    } else if (align === LEFT_ALIGN) {
        maxWidth -= minmaxLabelDim[0];
        x = -minmaxLabelDim[0];
    } else if (align === RIGHT_ALIGN) {
        maxWidth -= minmaxLabelDim[0];
    }

    return { maxWidth, maxHeight, x };
}

function getAxisBound(available, minmaxLabelDim, isHorizontal, orient, axisPadding) {
    let { maxWidth, maxHeight } = available;

    if (isHorizontal) {
        return available;
    }

    if (orient === RIGHT_ALIGN) {
        maxWidth -= minmaxLabelDim[0] + axisPadding;
    } else {
        maxWidth -= axisPadding;
    }

    return { maxWidth, maxHeight };
}

export { SizeSliderParser };
export { getMinmaxLabelBound };
