// @zohocharts/zc/src/legend/size-legend/slider/size-slider.js
//
import { Registry } from '../../../core/zc-registry';
import { LegendRegistry } from '../../legend-registry';
import { SizeLegend } from '../size-legend';
import { SizeSliderParser } from './size-slider-parser';
import { SliderAxisAdjustFactor } from './slider-axis-adjust';
import { MinmaxLabelAdjustFactor } from '../../minmax-label/legend-minmax-label-adjust';
import { HORIZONTAL_ALIGN, BOTTOM_ALIGN, LEFT_ALIGN, TOP_ALIGN } from '../../../core/align-constants';
import { mathMax, mathMin } from '../../../core/zc-constants';
import { bbox, perToPx, mergeJSON, each } from '../../../utils/zcutils';
import { getBrushScale, getInitialExtent } from '../../legend-brush';
import { getLegendAlignment } from '../../legend-utils';
import { getMaxToResize } from '../../legend-bound';
import { getSVGsymbol } from '../../../utils/svg-symbols';
import { getSliderMarkerPosition } from './size-slider-helpers';
import { getDefaultSliderOptions } from '../../../behaviours/visual-effects/brush-handles/brush-handle-options';
import { getHoverOptions, getDisabledOptions } from '../../events/legend-event-helpers';
import { getBoxSize } from '../../../view/list/list-helpers';

class SizeLegendSlider extends SizeLegend {
    constructor() {
        super(...arguments);
        this.brushOptions = this.getBrushOptions();
    }
    create() {
        let viewScaleComponent = this.viewScaleComponent();
        this.viewScale = viewScaleComponent.scale;
        this.brushOptions = this.getBrushOptions();

        let obj = { spec: this.spec, brushOptions: this.brushOptions, viewScaleComponent, align: () => this.align(this) };
        let mark = SizeSliderParser(obj);
        this._parser = mark._helper;
        return (this.encode = mark);
    }

    resize(width, height) {
        if (!width && !height) {
            return;
        }

        let // Element
            trackerEle = getEle(this, 'Tracker'),
            backgroundEle = getEle(this, 'LegendBackgroundSvg'),
            // BBox
            trackerBox = bbox(trackerEle),
            backgroundBox = bbox(backgroundEle),
            //
            { w, h } = getMaxToResize(width, height, this.spec.bound);

        w -= width ? mathMax(backgroundBox.width - trackerBox.width, 0) : 0;
        h -= height ? mathMax(backgroundBox.height - trackerBox.height, 0) : 0;

        this.updateScale(w, h);
        super.resize(w, h);
    }

    align(self) {
        let { alignSpec } = self._parser,
            { isHorizontal, padding, axisPadding, axisOrient, markerSize } = alignSpec,
            // Element
            titleEle = getEle(self, 'LegendTitleSvg'),
            trackerEle = getEle(self, 'Tracker'),
            sliderEle = getEle(self, 'SliderView'),
            axisEle = getEle(self, 'SliderAxis'),
            minmaxLabelEle = getEle(self, 'MinmaxLabel'),
            connectorEle = getEle(self, 'Connector'),
            selectionTrackerEle = getEle(self, 'SelectionTracker'),
            // BBox
            titleBox = bbox(titleEle),
            trackerBox = bbox(trackerEle),
            axisBox = bbox(axisEle),
            minmaxLabelsBox = bbox(minmaxLabelEle),
            //
            radius0 = markerSize[0][0] / 2,
            x,
            y;

        if (isHorizontal) {
            if (axisOrient === TOP_ALIGN) {
                x = radius0 - mathMin(0, axisBox.x);
                y = titleBox.height + axisBox.height + padding;

                translate([x, y + axisPadding - trackerBox.y], trackerEle, sliderEle, connectorEle, selectionTrackerEle);
                translate([x, y], axisEle, minmaxLabelEle);
            } else {
                x = radius0;
                y = titleBox.height + padding;

                translate([x, y - trackerBox.y], trackerEle, sliderEle, connectorEle, selectionTrackerEle);
                translate([x, y + trackerBox.height + axisPadding], axisEle, minmaxLabelEle);
            }
        } else {
            if (axisOrient === LEFT_ALIGN) {
                x = axisBox.width;
                y = titleBox.height + padding + mathMax(-mathMin(0, minmaxLabelsBox.y), radius0);

                translate([x + axisPadding, y], trackerEle, sliderEle, connectorEle, selectionTrackerEle);
                translate([x + trackerBox.x, y], axisEle, minmaxLabelEle);
            } else {
                x = 0;
                y = titleBox.height + padding + mathMax(-mathMin(0, minmaxLabelsBox.y), radius0);

                translate([x, y], trackerEle, sliderEle, connectorEle, selectionTrackerEle);
                translate([trackerBox.width + trackerBox.x + axisPadding, y], axisEle);
                translate([trackerBox.width + trackerBox.x, y], minmaxLabelEle);
            }
        }

        let titleX = mathMin(0, minmaxLabelEle.node().getBoundingClientRect().left - titleEle.node().getBoundingClientRect().left);
        translate([titleX, 0], titleEle);
    }

    getSliderArea() {
        // TODO: consider maxWidth, maxHeight, axis, minmaxlabels
        let scale = this.viewScale,
            { config, bound, area } = this.spec,
            { maxWidth, maxHeight } = bound,
            sliderConfig = config.slider,
            filterConfig = getDisabledOptions(config),
            scale_domain = scale.domain(),
            range = [scale(scale_domain[0]), scale(scale_domain[scale_domain.length - 1])],
            align = getLegendAlignment(config),
            isHorizontal = align.layout === HORIZONTAL_ALIGN,
            userWidth = filterConfig.type ? sliderConfig.width : isHorizontal ? sliderConfig.width : null,
            userHeight = filterConfig.type ? sliderConfig.height : isHorizontal ? null : sliderConfig.height,
            defaultWidth,
            defaultHeight,
            s1 = range[0] + range[range.length - 1],
            s2 = mathMax(range[0], range[range.length - 1]);

        if (isHorizontal) {
            defaultWidth = s1 * 1.5;
            defaultHeight = s2;
            maxWidth -= s1 / 2;
        } else {
            defaultWidth = s2;
            defaultHeight = s1 * 1.5;
            maxHeight -= s1 / 2;
        }

        let width = mathMin(perToPx(userWidth, area.width, defaultWidth), maxWidth),
            height = mathMin(perToPx(userHeight, area.height, defaultHeight), maxHeight);

        // let markerBound = getBoxSize(sliderConfig.marker, {}, align, scale, 0),
        //     d1 = [markerBound.width(scale_domain[0]), markerBound.height(scale_domain[0])],
        //     d2 = [markerBound.width(scale_domain[scale_domain.length - 1]), markerBound.height(scale_domain[scale_domain.length - 1])];
        //
        // if (isHorizontal) {
        //     width = mathMax(0, mathMin(width, maxWidth - d1[1] / 2 - d2[1] / 2));
        // } else {
        //     height = mathMax(0, mathMin(height, maxHeight - d1[1] / 2 - d2[1] / 2));
        // }

        return { width, height };
    }
    getBrushOptions() {
        this.viewScale = this.viewScale || this.viewScaleComponent().scale; // If legend is enabled, this.viewScale is needed for event api

        let dim = this.getSliderArea(),
            scale = this.viewScale,
            { config, format } = this.spec,
            { ranges, reversed } = config,
            { dataType } = this.spec.scaleComponent.schema.data,
            //
            align = getLegendAlignment(config),
            isHorizontal = align.layout === HORIZONTAL_ALIGN,
            //
            _range = isHorizontal ? [0, dim.width] : [0, dim.height],
            range = reversed ? _range.reverse() : _range,
            //
            brushScale = getBrushScale(config, dataType, scale, range),
            initialExtent = getInitialExtent(ranges, config.filter.range, scale, brushScale, dataType),
            //
            brushOptions = {
                config,
                reversed,
                ranges,
                range,
                isHorizontal,
                initialExtent,
                align,
                format,
                brushScale,
                scale: brushScale,
                sizeScale: scale,
                datatype: dataType,
                type: isHorizontal ? 'x' : 'y',
                orient: isHorizontal ? BOTTOM_ALIGN : LEFT_ALIGN,
                dimension: [dim.width, dim.height],
                scale_domain: scale.domain(),
                singleDomainContinuous: brushScale.singleDomainContinuous,
                markerBound: getBoxSize(config.slider.marker, {}, align, scale, 0),
                encode: 'z'
            };

        brushOptions.handles = getHandleOptions(brushOptions);

        return brushOptions;
    }
}

SizeLegendSlider.adjustFactor = (config, w, h, globalFont, size) => {
    let { maxWidth, maxHeight, slider: sliderConfig } = config,
        axisConfig = sliderConfig.axis,
        minmaxLabelsConfig = axisConfig.minmaxlabels,
        align = getLegendAlignment(config),
        isHorizontal = align.layout === HORIZONTAL_ALIGN,
        //
        axisAdjust = SliderAxisAdjustFactor(axisConfig, w, h, globalFont, align),
        minmaxLabelAdjust = MinmaxLabelAdjustFactor(minmaxLabelsConfig, w, h, globalFont, align),
        adjust = [0, 0];

    if (!maxHeight) {
        adjust[1] += axisAdjust[1] + minmaxLabelAdjust[1];

        if (!isHorizontal) {
            adjust[1] += size[0] / 2 + size[size.length - 1] / 2;
        }
    }

    if (!maxWidth) {
        adjust[0] += axisAdjust[0] + minmaxLabelAdjust[0];

        if (isHorizontal) {
            adjust[0] += size[0] / 2 + size[size.length - 1] / 2;
        }
    }

    return adjust;
};

function translate(xy, ...elements) {
    let _translate = `translate(${xy.join(' ')})`;
    each(elements, ele => ele.attr('transform', _translate));
}

function getEle(self, view) {
    let order = {
        LegendTitleSvg: 0,
        Tracker: 1,
        SelectionTracker: 2,
        SliderAxis: 3,
        Connector: 4,
        SliderView: 5,
        MinmaxLabel: 6
    };

    if (view === 'LegendBackgroundSvg') {
        return self.encode.children[0].getElement();
    }
    return self.encode.children[1].children[order[view]].getElement();
}

function getHandleOptions(options) {
    let { sizeScale, brushScale, config, dimension, align, isHorizontal } = options,
        bound = { width: dimension[0], height: dimension[1] },
        { slider: sliderConfig } = config,
        filterConfig = getDisabledOptions(config),
        markerConfig = sliderConfig.marker,
        hoverOptions = getHoverOptions(config),
        handleType = filterConfig.type,
        handleSymbol = markerConfig.shape || 'circle',
        shape,
        _transform;

    if (hoverOptions.type) {
        let op = getDefaultSliderOptions(dimension, isHorizontal);
        config.hoverOptions.slider = mergeJSON(op, hoverOptions, true);
    }

    if (handleType) {
        let op = getDefaultSliderOptions(dimension, isHorizontal);
        config.filter.slider = mergeJSON(op, filterConfig, true);
    }

    if (!handleType) {
        let position = getSliderMarkerPosition(sizeScale, brushScale, markerConfig, bound, align);
        shape = getSVGsymbol(handleSymbol);

        if (align.layout === HORIZONTAL_ALIGN) {
            _transform = d => [0, position.y(d)];
        } else {
            _transform = d => [position.x(d), 0];
        }

        return {
            shape,
            position: _transform,
            transform: d => `translate(${_transform(d).join(' ')}) `
        };
    }
}

SizeLegendSlider.encode = 'z';
SizeLegendSlider.view = 'slider';
SizeLegendSlider.domMode = 'svg';

export { SizeLegendSlider };
Registry.setComponent('sizeLegendSlider', SizeLegendSlider);
LegendRegistry.add(SizeLegendSlider);
