// @zohocharts/zc/src/legend/size-legend/slider/slider-connector-parser.js
//
import { PATH_ELEMENT } from '../../../core/element-constants';
import { HORIZONTAL_ALIGN } from '../../../core/align-constants';
import { CONNECTOR_LEGEND_CLASSNAME, CONNECTOR_LINE_CLASSNAME } from '../../legend-constants';
import { Mark } from '../../../mark/mark-utils';
import { mathMax } from '../../../core/zc-constants';
import { d3_selectAll } from '../../../core/d3-constants';
import { pick } from '../../../utils/zcutils';
import { createElement, elementToPolygonPoints } from '../../../utils/zcsvgutils';
import { getSVGsymbol } from '../../../utils/svg-symbols';
import { getSliderMarkerPosition } from './size-slider-helpers';

function ConnectorParser(spec) {
    let _helper = helper(spec);

    let groupMark = Mark('group', `.${CONNECTOR_LEGEND_CLASSNAME}`);

    let connectorMark = Mark('polygon', `.${CONNECTOR_LINE_CLASSNAME}`, _helper.data);
    connectorMark.config = _helper.config;

    groupMark.children = [connectorMark];
    return groupMark;
}

function helper(spec) {
    let { data, config } = spec,
        { connector: connectorConfig } = config,
        { show } = connectorConfig;

    return {
        data: show ? [data] : [],
        config: {
            points: getPoints(spec),
            ...connectorConfig
        }
    };
}

function getPoints(spec) {
    let { bound, scale, brushScale, align, config, filterConfig, markerBound, data: actData } = spec,
        { width, height } = markerBound,
        { marker: markerConfig } = config,
        shape = markerConfig.shape,
        handleType = filterConfig.type,
        isHorizontal = align.layout === HORIZONTAL_ALIGN,
        bandWidth = brushScale.bandwidth ? brushScale.bandwidth() : 0,
        position = getSliderMarkerPosition(scale, brushScale, markerConfig, bound, align),
        self = this, //TODO: check
        maxSize = mathMax(scale(actData[0]) / 2, scale(actData[actData.length - 1]) / 2);

    return (data, selection = []) => {
        let measure = data.map((d, i) => {
                let pos = pick(selection[i], brushScale(d) + bandWidth / 2),
                    r = scale(d) / 2;

                if (isHorizontal) {
                    return { x: pos, y: position.y(d), r, w: width(d), h: height(d) };
                } else {
                    return { x: position.x(d), y: pos, r, w: width(d), h: height(d) };
                }
            }),
            [m0, m1] = measure,
            points;

        if (handleType || shape === 'whisker' || shape === 'horizontal-whisker') {
            points = [[m0.x, m0.y], [m1.x, m1.y]];
        } else if (shape === 'circle') {
            if (isHorizontal) {
                points = [[m0.x + m0.r, m0.y], [m1.x - m1.r, m1.y]];
            } else {
                points = [[m0.x, m0.y + m0.r], [m1.x, m1.y - m1.r]];
            }
        } else if (shape === 'square' || shape === 'rect') {
            if (isHorizontal) {
                points = [[m0.x + m0.w / 2, m0.y], [m1.x - m1.w / 2, m1.y]];
            } else {
                points = [[m0.x, m0.y + m0.h / 2], [m1.x, m1.y - m1.h / 2]];
            }
        } else {
            let _node = d3_selectAll('svg.svgPattern');
            if (!_node.__cache__) {
                _node.__cache__ = _node.__cache__ || {};
                // TODO: work on this
                let tempEle = createElement(PATH_ELEMENT, '.temp_slider_marker', [1], _node);
                tempEle.attr('d', getSVGsymbol(shape)(maxSize * 2));

                _node.__cache__.edges = getEdges(tempEle);
                tempEle.remove();
            }

            let { right, left, top, bottom } = _node.__cache__.edges;

            if (isHorizontal) {
                let _right = [(right[0] / maxSize) * m0.r, (right[1] / maxSize) * m0.r],
                    _left = [(left[0] / maxSize) * m1.r, (left[1] / maxSize) * m1.r];

                points = [[m0.x + _right[0], m0.y + _right[1]], [m1.x + _left[0], m1.y + _left[1]]];
            } else {
                let _bottom = [(bottom[0] / maxSize) * m0.r, (bottom[1] / maxSize) * m0.r],
                    _top = [(top[0] / maxSize) * m1.r, (top[1] / maxSize) * m1.r];

                points = [[m0.x + _bottom[0], m0.y + _bottom[1]], [m1.x + _top[0], m1.y + _top[1]]];
            }
        }

        let isOverlap = isHorizontal
            ? points[0][0] > points[1][0] || points[1][0] < points[0][0]
            : points[0][1] > points[1][1] || points[1][1] < points[0][1];

        if (isOverlap) {
            // Is symbol overlap
            points = [];
        }

        return points;
    };
}

function getEdges(ele) {
    let polygonPoints = elementToPolygonPoints(ele),
        edge = {};

    polygonPoints.forEach(item => {
        let foundLeft = item[1] < 1 && item[1] > -1 && item[0] <= 0,
            foundRight = item[1] < 1 && item[1] > -1 && item[0] >= 0,
            foundTop = item[0] < 1 && item[0] > -1 && item[1] <= 0,
            foundBottom = item[0] < 1 && item[0] > -1 && item[1] >= 0;

        if (foundLeft) {
            if (!edge.left || edge.left[0] > item[0]) {
                edge.left = item;
            }
        }
        if (foundRight) {
            if (!edge.right || edge.right[0] < item[0]) {
                edge.right = item;
            }
        }
        if (foundTop) {
            if (!edge.top || edge.top[1] > item[1]) {
                edge.top = item;
            }
        }
        if (foundBottom) {
            if (!edge.bottom || edge.bottom[1] < item[1]) {
                edge.bottom = item;
            }
        }
    });

    return edge;
}

export { ConnectorParser, getPoints };
