// @zohocharts/zc/src/legend/size-legend/slider/slider-axis-parser.js
//
import { HORIZONTAL_ALIGN, RIGHT_ALIGN, LEFT_ALIGN, START_ALIGN, END_ALIGN, MIDDLE_ALIGN } from '../../../core/align-constants';
import { SLIDER_AXIS_CLASSNAME } from '../../legend-constants';
import { Mark } from '../../../mark/mark-utils';
import { d3axis } from '../../../axis/d3-axis';
import { d3_select } from '../../../core/d3-constants';
import { pick, setPropVal } from '../../../utils/zcutils';
import { addSVGtitleEle } from '../../../utils/zcsvgutils';
import { getSvgTransform } from '../../../utils/svg-transforms';
import { handleTicklabels } from '../nested/size-nested-ticklabel-handler';

function SliderAxisParser(spec) {
    let _helper = helper(spec),
        { show, axisSpec, trimText, updateAxisLine } = _helper,
        { tickValues } = axisSpec,
        { config } = spec;

    let groupMark = Mark('group', `.${SLIDER_AXIS_CLASSNAME}`);
    if (!show) {
        return groupMark;
    }

    let drawAxis = ele => {
        ele.call(d3axis, axisSpec, config);
        updateAxisLine(ele.selectAll('path.domain'));
        ele.selectAll('g.tick').sort((a, b) => (a === tickValues[0] || a === tickValues[tickValues.length - 1] ? -1 : 1));
        trimText(ele);

        let labelEle = getLabelEle(ele);
        labelEle.style('pointer-events', 'visible');
    };
    let axisGroupMark = Mark('group');
    axisGroupMark.onEnd = drawAxis;

    groupMark.children = [axisGroupMark];
    return groupMark;
}

function helper(spec) {
    let { scale, config, format, orient, viewBound, availableBound, align, data } = spec,
        { show, tickmark: tickMarkconfig, ticklabel: tickLabelConfig } = config,
        { tooltip, labels } = tickLabelConfig,
        isHorizontal = align.layout === HORIZONTAL_ALIGN,
        tickFormat = function(d, i) {
            let _text = labels && labels.length ? labels[i] : format(d);
            this.__formattedValue = _text;
            d3_select(this).text(_text);
        },
        //
        dummyChartObj = {
            dataObject: {},
            systemConf: { canvas: {}, chart: { plot: { border: {} } } }
        },
        dummyAxisObj = { ticks: { isHierarchialLevelType: () => {} } },
        //
        addTooltip = function() {
            addSVGtitleEle(d3_select(this), this.__formattedValue);
        },
        trimText = ele => {
            let labelEle = getLabelEle(ele),
                lineEle = getLineEle(ele),
                _anchor = {
                    anchor: getAnchor(orient),
                    translateFunction: ele => {
                        let textTranslate = getSvgTransform(ele).translate,
                            gTranslate = getSvgTransform(ele.parentNode).translate;

                        return [textTranslate[0] + gTranslate[0], textTranslate[1] + gTranslate[1]];
                    }
                };

            handleTicklabels(labelEle, lineEle, availableBound, viewBound, _anchor);

            //  ZC2424 after handle overflow, tooltip will not be shown
            if (tooltip === true) {
                labelEle.each(addTooltip);
            }
        };

    let axisSpec = {
        orient,
        scale,
        tickFormat,
        tickValues: data,
        tickSizeInner: pick(tickMarkconfig.size, 6),
        tickSizeOuter: 0,
        gridSize: null,
        tickPadding: pick(tickLabelConfig.margin, 3),
        rotation: 0,
        renderTicklabel: true,
        chart: dummyChartObj, // d3axis requires chartObj
        axis: dummyAxisObj // d3axis requires axisObj
    };

    return {
        show,
        axisSpec,
        trimText,
        updateAxisLine: ele => {
            if (scale.bandwidth) {
                let range = scale.range(),
                    bandWidth = scale.bandwidth(),
                    tickMarkStrokeWidth = pick(tickMarkconfig.strokeWidth, 1),
                    excess = bandWidth / 2 - tickMarkStrokeWidth / 2,
                    newRange = range[1] > range[0] ? [range[0] + excess, range[1] - excess] : [range[0] - excess, range[1] + excess],
                    path = isHorizontal ? `M${newRange[0]},0 L${newRange[1]},0` : `M0,${newRange[0]} L0,${newRange[1]}`;

                ele.attr('d', path);
            }
        }
    };
}

function getAnchor(orient) {
    return orient === RIGHT_ALIGN ? START_ALIGN : orient === LEFT_ALIGN ? END_ALIGN : MIDDLE_ALIGN;
}

function getLabelEle(ele) {
    return ele.selectAll('text.ticklabel');
}

function getLineEle(ele) {
    return ele.selectAll('line.minorTicks');
}

export { SliderAxisParser };
