// @zohocharts/zc/src/legend/size-legend/slider/tracker-parser.js
//
import { TRACK_LEGEND_CLASSNAME, TRACK_FLASH_CLASSNAME, TRACK_SYMBOL_CLASSNAME } from '../../legend-constants';
import { HORIZONTAL_ALIGN } from '../../../core/align-constants';
import { Mark } from '../../../mark/mark-utils';
import { d3_select } from '../../../core/d3-constants';
import { geometryUtils } from '../../../utils/geometry-utils';
import { elementToPolygonPoints } from '../../../utils/zcsvgutils';
import { getSliderMarkerPosition } from './size-slider-helpers';

function TrackerParser(spec) {
    let _helper = helper(spec);

    let groupMark = Mark('group', `.${_helper.className}`);
    groupMark.config = {
        visibility: _helper.show ? null : 'hidden',
        'shape-rendering': 'crispedges' // ZC2552
    };

    let symbolMark = SymbolParser(_helper);
    let flashMark = FlashParser(_helper);

    groupMark.children = [symbolMark, flashMark];
    return groupMark;
}

function FlashParser(spec) {
    let { flashSpec, styles } = spec;

    return {
        mark: 'polygon',
        selector: `.${TRACK_FLASH_CLASSNAME}`,
        config: {
            points: flashSpec.points,
            curve: 'linearClosed',
            ...styles
        },
        onStart: flashSpec.onStart
    };
}

function SymbolParser(spec) {
    let { symbolSpec, styles } = spec;

    let symbolMark = Mark('point', `.${TRACK_SYMBOL_CLASSNAME}`, symbolSpec.data);
    symbolMark.config = { ...symbolSpec, ...styles };
    return symbolMark;
}

function helper(spec) {
    let { bound, data, scale, brushScale, config, filterConfig, align, markerBound, trackConfig, className } = spec,
        { marker: markerConfig } = config,
        handleType = filterConfig.type,
        position = getSliderMarkerPosition(scale, brushScale, markerConfig, bound, align),
        { x, y } = position,
        symbolEle;

    let flashSpec = {
            points: () => getPolygon(spec, position, symbolEle),
            onStart: ele => (symbolEle = d3_select(ele.node().parentNode).selectAll(`path.${TRACK_SYMBOL_CLASSNAME}`)) // TODO: pass symbol elements here
        },
        symbolSpec = {
            data: handleType ? [] : data,
            x: d => x(d),
            y: d => y(d),
            shape: markerConfig.shape,
            size: d => [markerBound.width(d), markerBound.height(d)] // TODO: size should handle array, number, function
        };

    return {
        show: trackConfig.show,
        flashSpec,
        symbolSpec,
        styles: trackConfig,
        className: className || TRACK_LEGEND_CLASSNAME
    };
}

function getPolygon(spec, position, symbolEle) {
    let { data, config, filterConfig, align, markerBound } = spec,
        { width, height } = config,
        shape = config.marker.shape,
        handleType = filterConfig.type,
        isHorizontal = align.layout === HORIZONTAL_ALIGN,
        { x, y, r } = position,
        measure = data.map(d => {
            return { x: x(d), y: y(d), r: r(d) / 2, w: markerBound.width(d), h: markerBound.height(d) };
        }),
        tangent;

    if (handleType) {
        let d1 = measure[0],
            d2 = measure[1],
            size = (isHorizontal ? height : width) || 0,
            d1_r = size / 2 || d1.r,
            d2_r = size / 2 || d2.r;

        tangent = isHorizontal
            ? [
                  [d1.x, d1.y + d1_r],
                  [d2.x, d2.y + d2_r],
                  [d1.x, d1.y - d1_r],
                  [d2.x, d2.y - d2_r]
              ]
            : [
                  [d1.x + d1_r, d1.y],
                  [d2.x + d2_r, d2.y],
                  [d1.x - d1_r, d1.y],
                  [d2.x - d2_r, d2.y]
              ];
    } else if (shape === 'circle') {
        tangent = geometryUtils.outerTangentPoints$Circle(measure[0], measure[1]);
    } else if (shape === 'square' || shape === 'rect') {
        let d1 = measure[0],
            d2 = measure[1];
        tangent = isHorizontal
            ? [
                  [d1.x + d1.w / 2, d1.y - d1.h / 2],
                  [d2.x - d2.w / 2, d2.y - d2.h / 2],
                  [d1.x + d1.w / 2, d1.y + d1.h / 2],
                  [d2.x - d2.w / 2, d2.y + d2.h / 2]
              ]
            : [
                  [d1.x + d1.w / 2, d1.y + d1.h / 2],
                  [d2.x + d2.w / 2, d2.y - d2.h / 2],
                  [d1.x - d1.w / 2, d1.y + d1.h / 2],
                  [d2.x - d2.w / 2, d2.y - d2.h / 2]
              ];
    } else if (shape === 'horizontal-whisker' || shape === 'whisker') {
        let d1 = measure[0],
            d2 = measure[1];
        tangent = isHorizontal
            ? [
                  [d1.x, d1.y - d1.r],
                  [d2.x, d2.y - d2.r],
                  [d1.x, d1.y + d1.r],
                  [d2.x, d2.y + d2.r]
              ]
            : [
                  [d1.x + d1.r, d1.y],
                  [d2.x + d2.r, d2.y],
                  [d1.x - d1.r, d1.y],
                  [d2.x - d2.r, d2.y]
              ];
    } else {
        let polygonPoints = [];
        symbolEle.each(function() {
            polygonPoints.push(elementToPolygonPoints(this));
        });

        tangent = geometryUtils.tangentPoints$Polygon(polygonPoints, isHorizontal, measure);
    }

    return tangent && tangent.length ? [tangent[0], tangent[1], tangent[3], tangent[2]] : [];
}

export { TrackerParser, helper };
