// @zohocharts/zc/src/legend/minmax-label/legend-minmax-label-parser.js
//
import { MINMAXLABEL_LEGEND_CLASSNAME } from '../legend-constants';
import { TEXT_ELEMENT } from '../../core/element-constants';
import { LEFT_ALIGN, TOP_ALIGN, BOTTOM_ALIGN, RIGHT_ALIGN } from '../../core/align-constants';
import { Registry } from '../../core/zc-registry';
import { Mark } from '../../mark/mark-utils';
import { isArray } from '../../utils/zcutils';
import { mathMax } from '../../core/zc-constants';

function MinmaxLabelParser(spec) {
    let _helper = helper(spec),
        { data } = spec,
        { show, text, x, y, dy, anchor, styles, trimText } = _helper;

    let groupMark = Mark('group', `.${MINMAXLABEL_LEGEND_CLASSNAME}`);
    let labelMark = Mark(TEXT_ELEMENT, undefined, show ? data : []);
    labelMark.config = { ...styles, text, x, y, dy, anchor };
    labelMark.onEnd = trimText;

    groupMark.children = [labelMark];
    return groupMark;
}

function helper(spec) {
    let {
            config,
            format,
            data,
            trimText,
            index = i => i,
            orient = RIGHT_ALIGN,
            dimension,
            sliderSize = [
                [0, 0],
                [0, 0]
            ]
        } = spec,
        { show, align, padding = 2, text: labels } = config,
        position = data.map((d, i) => getPositionInfo(orient, dimension, align, padding, sliderSize, index(i)));

    return {
        show,
        trimText,
        text: (d, i) => (labels ? labels[index(i)] : format(d, index(i))),
        x: (d, i) => position[index(i)].x,
        y: (d, i) => position[index(i)].y,
        anchor: (d, i) => position[index(i)].anchor,
        dy: (d, i) => position[index(i)].dy,
        styles: config // TODO: pick font properties alone
    };
}

function getPositionInfo(orient, dimension, align, padding, sliderSizes, i) {
    let sliderSize = isArray(sliderSizes[i]) ? sliderSizes[i] : sliderSizes;
    var x, y, dy, ANCHOR;
    if (orient === BOTTOM_ALIGN) {
        if (align === TOP_ALIGN) {
            dy = '-0.3em';
            ANCHOR = i === 0 ? null : 'end';
            y = -dimension[1] - padding - mathMax(0, (sliderSize[1] - dimension[1]) / 2);
            if (i === 0) {
                x = 0;
            } else {
                x = dimension[0];
            }
        } else if (align === BOTTOM_ALIGN) {
            dy = '1em';
            ANCHOR = i === 0 ? null : 'end';
            y = padding + mathMax(0, (sliderSize[1] - dimension[1]) / 2);
            if (i === 0) {
                x = 0;
            } else {
                x = dimension[0];
            }
        } else {
            dy = '0.3em';
            ANCHOR = i === 0 ? 'end' : null;
            y = -dimension[1] / 2;
            if (i === 0) {
                x = -padding - sliderSize[0] / 2;
            } else {
                x = dimension[0] + padding + sliderSize[0] / 2;
            }
        }
    } else if (orient === TOP_ALIGN) {
        if (align === TOP_ALIGN) {
            dy = '-0.3em';
            ANCHOR = i === 0 ? null : 'end';
            y = -padding - mathMax(0, (sliderSize[1] - dimension[1]) / 2);
            if (i === 0) {
                x = 0;
            } else {
                x = dimension[0];
            }
        } else if (align === BOTTOM_ALIGN) {
            dy = '1em';
            ANCHOR = i === 0 ? null : 'end';
            y = dimension[1] + padding + mathMax(0, (sliderSize[1] - dimension[1]) / 2);
            if (i === 0) {
                x = 0;
            } else {
                x = dimension[0];
            }
        } else {
            dy = '0.3em';
            ANCHOR = i === 0 ? 'end' : null;
            y = dimension[1] / 2;
            if (i === 0) {
                x = -padding - sliderSize[0] / 2;
            } else {
                x = dimension[0] + padding + sliderSize[0] / 2;
            }
        }
    } else if (orient === RIGHT_ALIGN) {
        if (align === LEFT_ALIGN) {
            dy = i !== 0 ? '0.71em' : '0em';
            ANCHOR = 'end';
            x = -dimension[0] - padding - mathMax(0, (sliderSize[1] - dimension[0]) / 2);
            if (i !== 0) {
                y = 0;
            } else {
                y = dimension[1];
            }
        } else if (align === RIGHT_ALIGN) {
            dy = i !== 0 ? '0.71em' : '0em';
            ANCHOR = null;
            x = padding + mathMax(0, (sliderSize[1] - dimension[0]) / 2);
            if (i !== 0) {
                y = 0;
            } else {
                y = dimension[1];
            }
        } else {
            dy = i !== 0 ? '-0.3em' : '1em';
            ANCHOR = 'middle';
            x = -dimension[0] / 2;
            if (i !== 0) {
                y = -padding - sliderSize[0] / 2;
            } else {
                y = dimension[1] + padding + sliderSize[0] / 2;
            }
        }
    } else {
        if (align === LEFT_ALIGN) {
            dy = i !== 0 ? '0.71em' : '0em';
            ANCHOR = 'end';
            x = -padding - mathMax(0, (sliderSize[1] - dimension[0]) / 2);
            if (i !== 0) {
                y = 0;
            } else {
                y = dimension[1];
            }
        } else if (align === RIGHT_ALIGN) {
            dy = i !== 0 ? '0.71em' : '0em';
            ANCHOR = null;
            x = dimension[0] + padding + mathMax(0, (sliderSize[1] - dimension[0]) / 2);
            if (i !== 0) {
                y = 0;
            } else {
                y = dimension[1];
            }
        } else {
            dy = i !== 0 ? '-0.3em' : '1em';
            ANCHOR = 'middle';
            x = dimension[0] / 2;
            if (i !== 0) {
                y = -padding - sliderSize[0] / 2;
            } else {
                y = dimension[1] + padding + sliderSize[0] / 2;
            }
        }
    }
    return {
        x: x,
        y: y,
        dy: dy,
        anchor: ANCHOR
    };
}

export { MinmaxLabelParser };
Registry.setComponent('minmaxLabel.parser', MinmaxLabelParser);
