// @zohocharts/zc/src/polaraxis/parser/config/polaraxis-events.js
import { d3_event } from "../../../core/d3-constants";
import { invokeFunction, mergeJSON, validateNullVal } from "../../../utils/zcutils";

let columnInfo = {};
export function _polarAxisEvents(data) {
    const self = this;

    const chart = this.chart,
        renderer = this.renderer,
        ticklabel = data.ticklabel,
        events_tick = ticklabel.events,
        columns = chart.systemConf.metadata.columns,
        columnIndices = renderer.columnIndices();

    columnInfo.data = columnIndices.map(function (d) {
        return mergeJSON(columns[d], { columnindex: d }, true);
    })

    let eventDisabled, mousemoveFn, mouseoutFn, clickFn, dblclickFn;
    if (events_tick != null) {
        eventDisabled = validateNullVal(events_tick.enabled, true) == false;
        mousemoveFn = events_tick.mousemove;
        mouseoutFn = events_tick.mouseout;
        clickFn = events_tick.click;
        dblclickFn = events_tick.doubleclick;
    }

    const mousemove = function (e, d) {
        callback(mousemoveFn, d3_event(e), d);
    },
        mouseout = function (e, d) {
            callback(mouseoutFn, d3_event(e), d);
        },
        click = function (e, d) {
            callback(clickFn, d3_event(e), d);
        },
        dblclick = function (e, d) {
            callback(dblclickFn, d3_event(e), d);
        };

    const mouseHandlers = {
        mousemove: mousemove,
        mouseout: mouseout,
        click: click,
        dblclick: dblclick
    },
        touchHandlers = {
            singletouchmove: mousemove,
            touchend: mouseout,
            tap: click,
            dbltap: dblclick
        }

    return { mouse: mouseHandlers, touch: touchHandlers };
}

function callback(func, e, d) {
    if (func != null) {
        columnInfo.data = columnInfo.data.map(function (dt) {
            dt.value = d;
            return dt;
        });
        invokeFunction(func, e, columnInfo, chartObj);
    }
}
