// @zohocharts/zc/src/label/parser/boundBox-marks.js
import { getChartTypeByIndex } from '../../core/charttype-helpers';
import { isSunburstChartType } from '../../events/events-utils';
import { setOpacityForTickFilterElements } from '../../plot/svg/set-pivot-attr';
import { pick } from '../../utils/zcutils';
import { findQueue } from '../labelrenderer';
import { DATALABELS_LABELTYPE, INNERLABEL_LABELTYPE, PARENTLABELS_LABELTYPE } from '../labeltype-constants';

export function getBoundBoxParser(data, renderOnInit, anime_config, parent) {
    var _this = this,
        enterMarks,
        updateMarks,
        exitMarks,
        staticSVG = _this.chart.staticPaneSVG,
        isParentLabel = _this.datalabel_type === PARENTLABELS_LABELTYPE,
        isInnerLabel = _this.isInnerLabel,
        label_type = isParentLabel ? 'parent' : isInnerLabel ? 'inner' : '',
        _Bb = staticSVG.selectAll(
            `.${label_type}boundboxseries_${pick(this.conf.rendererIndex() || this.conf.datasetIndex() || _this.index)}`
        ), //0 - for funnel chart nd some chart type cases
        parent,
        showBoundBox,
        labelQueue,
        anim_dur = anime_config && anime_config._duration || 100,//for other cases other than bar
        anim_delay = anime_config && anime_config._delay || 0,
        racing = this.chart.racing && this.chart.racing.syncTransition;
    if (!_Bb.node()) {
        return;
    }
    _Bb.each(d => {
        showBoundBox = d.showBoundBox;
        labelQueue = d.labelQueue;
    });
    var paneid = labelQueue.paneid,
        innerLabels = /(pieInnerLabels|funnelInnerLabels)/.test(labelQueue.charttype),
        legendOrder = labelQueue.args[1],
        labelType =
            labelQueue.charttype === 'pieInnerLabels' &&
            isSunburstChartType(getChartTypeByIndex(_this.chart, paneid, legendOrder).value) &&
            labelQueue.args[labelQueue.args.length - 1]
                ? PARENTLABELS_LABELTYPE
                : innerLabels
                ? INNERLABEL_LABELTYPE
                : DATALABELS_LABELTYPE,
        args = {
            legendOrder: legendOrder,
            rendererIndex: labelQueue.rendererIndex,
            paneid: paneid,
            labelType: labelType
        };

    if (_this.properties && _this.properties.rangeChart) {
        _Bb = _Bb.selectAll(`.rangePos_${_this.properties.yrangePos}`);
    }
    parent = _Bb.selectAll(`.subseries_${_this.subseriesIdx}`);

    var isOrdinal = this.chart.dataset.getDataType("x",0) === 'ordinal';
    var boundBoxMarks = {
        name: 'rect',
        selector: '.zc_label_background',
        mark: 'path',
        order: '.customLabel',
        data: showBoundBox ? data : [],
        dataAccessor: racing && isOrdinal ? (_)=>this.chart.dataset.getX(_) : null,
        parent: parent,
        config: {
            strokeWidth: _this.helpers.styles.boundBox.strokeWidth,
            dashStyle: _this.helpers.styles.boundBox['stroke-dasharray'],
            lineCap: _this.helpers.styles.boundBox['stroke-linecap']
        },
        properties: {
            styles: {
                fill: (d, i) => _this.helpers.styles.boundBox.fill(d, i, args),
                'fill-opacity': _this.helpers.styles.boundBox['fill-opacity'],
                stroke: (d, i) => _this.helpers.styles.boundBox.stroke(d, i, args),
                'stroke-opacity': _this.helpers.styles.boundBox['stroke-opacity'],
                // 'stroke-dasharray': _this.helpers.styles.boundBox['stroke-dasharray'],
                // 'stroke-linecap': _this.helpers.styles.boundBox['stroke-linecap'],
                transition: _this.helpers.styles.boundBox.transition,
                opacity: (d, i) => _this.helpers.styles.boundBox.opacity(d, i, args)
            },
            attrs: {
                d: (d, i) => _this.helpers.styles.boundBox.d(d, i, args),
                id: (d, i) => _this.helpers.styles.boundBox.id(d, i, args),
                levelbycolumn: (d, i) => _this.helpers.styles.boundBox.levelbycolumn(d, i, args),
                levelbyrow: (d, i) => _this.helpers.styles.boundBox.levelbyrow(d, i, args),
                display: (d, i) => _this.helpers.styles.boundBox.display(d, i, args)
            }
        },
        onEnd: function(d) {
            setOpacityForTickFilterElements(d, _this.chart, 'label', _this.actualData);
        }
    };

    if (racing) {
        var raceBoundBox_position = _this.helpers.position.racingBoundBox;
        enterMarks = {
            transform: (d, k) => {
                return raceBoundBox_position.start(d, k);
            }
        };
        updateMarks = [
            {
                transition: racing,
                transform: {
                    property: 'attrTween',
                    to: function(d, i) {
                        let interpolator = d3.interpolateString(
                            this.getAttribute('transform'),
                            raceBoundBox_position.end(d, i)
                        );
                        return function(t) {
                            return interpolator(t);
                        };
                    }
                }
            }
        ];
        exitMarks = {
            transition: racing,
            transform: function(d, k) {
                return raceBoundBox_position.exit(d, k, this);
            }
        }
        boundBoxMarks.config.enter = enterMarks;
        boundBoxMarks.anime = {};
        boundBoxMarks.anime.update = updateMarks;
        boundBoxMarks.anime.exit = exitMarks;
    } else if (renderOnInit) {
        enterMarks = {
            transform: (d, k) => {
                return _this.helpers.position.boundBox.start.transform(d, k);
            }
        };
        updateMarks = [
            {
                duration: this.isRedraw == 'redraw' ? anim_dur : 0,
                delay: anim_delay,
                transform: {
                    property: 'attrTween',
                    to: function(d, i) {
                        let interpolator = d3.interpolateString(
                            this.getAttribute('transform'),
                            _this.helpers.position.boundBox.stop_1.transform(d, i, this)
                        );
                        return function(t) {
                            return interpolator(t);
                        };
                    }
                }
            },
            {
                duration: anim_dur,
                delay: anim_delay,
                transform: {
                    property: 'attrTween',
                    to: function(d, i) {
                        let interpolator = d3.interpolateString(
                            this.getAttribute('transform'),
                            _this.helpers.position.boundBox.end.transform(d, i)
                        );
                        return function(t) {
                            return interpolator(t);
                        };
                    }
                }
            }
        ];
        boundBoxMarks.config.enter = enterMarks;
        boundBoxMarks.anime = {};
        boundBoxMarks.anime.update = updateMarks;
    } else {
        enterMarks = {
            transform: (d, k) => {
                return _this.helpers.position.boundBox.end.transform(d, k);
            }
        };
        updateMarks = {
            duration: 0,
            transform: (d, k) => {
                return _this.helpers.position.boundBox.end.transform(d, k);
            }
        };
        exitMarks = {
            duration: 0,
            delay: 0
        };
        boundBoxMarks.config.enter = enterMarks;
        boundBoxMarks.config.update = updateMarks;
        boundBoxMarks.config.exit = exitMarks;
    }

    return boundBoxMarks;
}
