// @zohocharts/zc/src/label/parser/datalabel-encode.js
import { colorscale_helpers } from '../../axis/colorscale-helpers';
import { LEFT_ALIGN, RIGHT_ALIGN } from '../../core/align-constants';
import { d3_scaleLinear } from '../../core/d3-constants';
import { charttype } from '../../core/main-zc-init';
import { NONE, NULL } from '../../core/zc-constants';
import { Registry } from '../../core/zc-registry';
import { effects } from '../../effects/zceffects';
import { getOpacityForTickFilterElements } from '../../plot/svg/set-pivot-attr';
import { getBoundBoxPathDescription } from '../../tooltip/svg-tooltip-pathParser';
import { getDashArray, getLineCap } from '../../utils/dashstyles';
import { getFontSizeByHierarchy } from '../../utils/zcdatautils';
import {
    arrayIncludes,
    bbox,
    defined,
    getBorderRadius,
    getContrastColor,
    invokeFunction,
    isUndefined
} from '../../utils/zcutils';
import {
    getBoundBoxColor,
    getBoundBoxHookConf,
    getBoundBoxPadding,
    getMergedBoundBoxOptions,
    getMetaConfigForBoundbox,
    hasBoundBox
} from '../boundBoxUtils';
import { DATALABELS_LABELTYPE } from '../labeltype-constants';

export function datalabel_style() {
    const KEY = 'label';
    var _this = this,
        { chart, properties, dataLevelFontColor, dataLevelStrokeColor } = _this,
        _sereisInfo = _this.subSeries;

    _this.helpers.styles[KEY] = {
        fill: function(d, i) {
            if (this.ctype === charttype.waterfall && dataLabels.fontColor !== undefined && dataLabels.fontColor == NULL) {
                var color = colorscale_helpers.getWaterfallColor(d, properties.rp.categoryColorConfig);
                if (color) {
                    return color;
                }
            }
            var y = dataLevelFontColor;
            return y && y(d, i);
        },
        display: function(d, i) {
            return _this.subSeries[i].display === false ? NONE : null;
        },
        opacity: function(d, i) {
            if (chart.dataObject.hierarchical && chart.cache.hierarchical) {
                return getOpacityForTickFilterElements(d, i, chart, 'label', _this.actualData);
            }
        },
        'stroke-width': function(data, i, isOutlineText) {
            if (isOutlineText) {
                return _this.properties.dataLabels.strokeWidth;
            }
        },
        'filtered-opacity': function(d, i) {
            if (chart.dataObject.hierarchical && chart.cache.hierarchical) {
                return getOpacityForTickFilterElements(d, i, chart, 'label', _this.actualData);
            }
        },
        dy: function(d, i) {
            return _this.subSeries[i].DY;
        },
        'text-anchor': function(d, i) {
            return _this.subSeries[i].ANCHOR;
        },
        id: function(d, i) {
            return _this.subSeries[i].ID;
        },
        'clip-path': function(d, i) {
            var bb = bbox(this);
            var properties = {
                    name: 'textClip',
                    id: chart.id,
                    x: bb.x,
                    y: bb.y,
                    width: _this.subSeries[i].width,
                    height: bb.height
                },
                args = {
                    chart: chart
                };
            var clipID = effects.get(effects.svg_clips, 'rect', properties, args);
            return `url(${chart.defsLocation}#${clipID})`;
        },
        levelbycolumn: function(_d, i) {
            let d = _this.actualData ? _this.actualData[_this.actualData.length - 1 - i] : _d.data || _d;
            if (d.levelByColumn) {
                return d.levelByColumn;
            }
        },
        levelbyrow: function(_d, i) {
            let d = _this.actualData ? _this.actualData[_this.actualData.length - 1 - i] : _d.data || _d;
            if (d.levelByRow) {
                return d.levelByRow;
            }
        },
        stroke: function(needStroke) {
            var strokeColor = _this.properties.dataLabels.strokeColor;
            if (needStroke) {
                // apply datalabels contrast color
                if (strokeColor === 'contrast') {
                    return function(d, i) {
                        var clr = dataLevelFontColor ? dataLevelFontColor(d, i) : _this.datalabelOptions.fill;
                        return getContrastColor(clr);
                    };
                } else {
                    return _this.dataLevelStrokeColor;
                }
            }
        }
    };

    _this.helpers.text[KEY] = function(d, i) {
        var str = _this.subSeries[i].text;
        return str;
    };
}
export function datalabel_position() {
    const KEY = 'label';
    var _this = this,
        _sereisInfo = _this.subSeries,
        chart = _this.chart,
        properties = _this.properties,
        rotation = properties.rotation;

    _this.helpers.position[KEY] = {};

    _this.helpers.position[KEY].start = {};
    _this.helpers.position[KEY].stop_1 = {};
    _this.helpers.position[KEY].end = {};

    var isMultiLineAnimation = (_this.properties.isMultiLine && _this.animate) || _this.animate; //if renderOnInit is true, it means animation is enabled.
    //if isMultiLineAnimation is true - translate will be set to transform to position the label, and all x,y will be set to 0

    _this.helpers.position[KEY].start.transform = (d, i) => {
        var val = '',
            _s = _this.subSeries[i],
            rotation = properties.rotation,
            pos = _s.xy || [_s.X, _s.Y],
            plotH = chart.plotarea.height;

        if (chart.axes.rotated) {
            pos = (_s.xy && [0, _s.xy[1]]) || [0, _s.Y];
        } else {
            pos = (_s.xy && [_s.xy[0], plotH]) || [_s.X, plotH];
        }
        if (_s.xy) {
            val = `translate(${pos})`;
        }
        if (isMultiLineAnimation) {
            if (defined(rotation)) {
                val = `translate(${pos}) rotate(${rotation})`;
            } else {
                val = `translate(${pos})`;
            }
        } else if (defined(rotation)) {
            val += `rotate(${rotation} ${pos})`;
        }
        return val || NULL;
    };
    _this.helpers.position[KEY].start.x = (d, i) => {
        var _s = _this.subSeries[i];
        if (isMultiLineAnimation) {
            return 0;
        }
        if (chart.axes.rotated) {
            return 0;
        }
        return (_s.xy && _s.xy[0]) || _s.X;
    };
    _this.helpers.position[KEY].start.y = (d, i) => {
        var _s = _this.subSeries[i];
        if (isMultiLineAnimation) {
            return 0;
        }
        if (chart.axes.rotated) {
            return (_s.xy && _s.xy[1]) || _s.Y;
        }
        return chart.plotarea.height;
    };

    _this.helpers.position[KEY].end.transform = (d, i) => {
        var val = '',
            rotation = properties.rotation,
            _s = _this.subSeries[i],
            pos = _s.xy || [_s.X, _s.Y],
            _minDist,
            isRotation_pos;

        if (defined(rotation)) {
            (_minDist = 360 - rotation), (isRotation_pos = rotation > 0);
            rotation = _minDist < rotation ? (isRotation_pos ? -_minDist : _minDist) : rotation;
        }

        if (_s.xy) {
            val = `translate(${pos})`;
        }
        if (isMultiLineAnimation) {
            if (defined(rotation)) {
                val = `translate(${pos}) rotate(${rotation})`;
            } else {
                val = `translate(${pos})`;
            }
        } else if (defined(rotation)) {
            val += `rotate(${rotation} ${pos})`;
        }

        return val || NULL;
    };
    _this.helpers.position[KEY].end.x = (d, i) => {
        var _s = _this.subSeries[i],
            rotation = properties.rotation;
        if (isMultiLineAnimation) {
            return 0;
        }
        return _s.X;
    };
    _this.helpers.position[KEY].end.y = (d, i) => {
        var _s = _this.subSeries[i],
            rotation = properties.rotation;
        if (isMultiLineAnimation) {
            return 0;
        }
        return _s.Y;
    };
    _this.helpers.position[KEY].end.translateXY = (d, i) => {
        var _s = _this.subSeries[i],
            val = _s._translate;
       
        return val || NULL;
    };

    _this.helpers.position[KEY].stop_1.transform = (d, i, ele) => {
        var _s = _this.subSeries[i],
            val = '',
            eleInfo = $ZC.get('domUtils.getSvgTransform')(ele),
            pos = _s.xy || [_s.X, _s.Y],
            _rot = eleInfo.rotate;

        if (chart.axes.rotated) {
            var _x = Number(ele.getAttribute('x')) || eleInfo.translate[0];
            if (_s.xy) {
                val = `translate(${[_x, pos[1]]})`;
            }
            if (isMultiLineAnimation) {
                if (defined(rotation)) {
                    val = `translate(${[_x, pos[1]]}) rotate(${rotation})`;
                } else {
                    val = `translate(${[_x, pos[1]]})`;
                }
            } else if (defined(_rot)) {
                val += `rotate(${_rot} ${[_x, pos[1]]})`;
            }
        } else {
            var _y = Number(ele.getAttribute('y')) || eleInfo.translate[1];
            if (_s.xy) {
                val = `translate(${[pos[0], _y]})`;
            }
            if (isMultiLineAnimation) {
                if (defined(rotation)) {
                    val = `translate(${[pos[0], _y]}) rotate(${rotation})`;
                } else {
                    val = `translate(${[pos[0], _y]})`;
                }
            } else if (defined(_rot)) {
                val += `rotate(${_rot} ${[pos[0], _y]})`;
            }
        }
        return val || NULL;
    };
    _this.helpers.position[KEY].stop_1.x = (d, i, ele) => {
        var _s = _this.subSeries[i],
            pos = _s.X,
            eleInfo = $ZC.get('domUtils.getSvgTransform')(ele),
            _rot = eleInfo.rotate;

        if (isMultiLineAnimation) {
            return 0;
        }

        if (this.chart.axes.rotated) {
            return ele.getAttribute('x');
        } else {
            return pos;
        }
    };
    _this.helpers.position[KEY].stop_1.y = (d, i, ele) => {
        var _s = _this.subSeries[i],
            pos = _s.Y,
            eleInfo = $ZC.get('domUtils.getSvgTransform')(ele),
            _rot = eleInfo.rotate;

        if (isMultiLineAnimation) {
            return 0;
        }
        if (this.chart.axes.rotated) {
            return pos;
        } else {
            return ele.getAttribute('y');
        }
    };

    _this.helpers.position[KEY].dy = (d, i) => {
        var _s = _this.subSeries[i];
        return _s.DY;
    };
}
export function racingLabel_position() {
    const KEY = 'racingLabel';
    var _this = this,
        _sereisInfo = _this.subSeries,
        chart = _this.chart,
        properties = _this.properties;

    _this.helpers.position[KEY] = {};

    _this.helpers.position[KEY].start = {};
    _this.helpers.position[KEY].exit = {};
    _this.helpers.position[KEY].end = {};

    _this.helpers.position[KEY].end.transform = (d, i, ele) => {
        var _s = _this.subSeries[i],
            pos = _s.xy || [_s.X, _s.Y],
            rotation = properties.rotation,
            val = '',
            _minDist,
            isRotation_pos;

        if (defined(rotation)) {
            (_minDist = 360 - rotation), (isRotation_pos = rotation > 0);
            rotation = _minDist < rotation ? (isRotation_pos ? -_minDist : _minDist) : rotation;
        }
        var race = _this.chart.racing,
            jumpDuration = race.jumpDuration;

        if (defined(jumpDuration)) {
            var from_ele_info = Registry.getComponent('domUtils.getSvgTransform')(ele),
                from_x = from_ele_info.translate[0],
                from_y = from_ele_info.translate[1],
                from_rotate = from_ele_info.rotate,
                to_x = pos[0],
                to_y = pos[1],
                to_rotate = rotation,
                x = d3.interpolate(from_x, to_x),
                y = d3.interpolate(from_y, to_y),
                rotate = d3.interpolate(from_rotate, to_rotate);

            return {
                x,
                y,
                rotate
            };
        } else {
            val = `translate(${pos})`;
            if (defined(rotation)) {
                val = `translate(${pos}) rotate(${rotation})`;
            }
            return val;
        }
    };

    _this.helpers.position[KEY].start.transform = (d, i) => {
        var _s = _this.subSeries[i],
            pos = _s.xy || [_s.X, _s.Y],
            rotation = properties.rotation,
            val = '',
            plotH = chart.plotarea.height,
            yOrient = this.chart.axes.y.groups[0].prerenderComp.orient,
            bandWidth = _this.scales.x.bandwidth && _this.scales.x.bandwidth() || 0;
    
        if (this.chart.axes.rotated && yOrient === "bottom") {
                val = `translate(${[pos[0], -this.chart.plotarea.top + bandWidth/2]})`;
                if (defined(rotation)) {
                    val = `translate(${[pos[0], -this.chart.plotarea.top + bandWidth/2]}) rotate(${rotation})`;
                }
                return val;
        }
        val = `translate(${[pos[0], plotH + 20 + bandWidth/2]})`;
        if (defined(rotation)) {
            val = `translate(${[pos[0], plotH + 20 + bandWidth/2]}) rotate(${rotation})`;
        }
        return val;
    };

    _this.helpers.position[KEY].exit.transform = (d, i, ele) => {
        var eleInfo = $ZC.get('domUtils.getSvgTransform')(ele),
            plotH = chart.plotarea.height,
            _rot = eleInfo.rotate,
            transX = eleInfo.translate[0],
            yOrient = this.chart.axes.y.groups[0].prerenderComp.orient,
            bandWidth = _this.scales.x.bandwidth && _this.scales.x.bandwidth() || 0;

        var race = _this.chart.racing,
            jumpDuration = race.jumpDuration;

        if (defined(jumpDuration)) {
            var from_ele_info = Registry.getComponent('domUtils.getSvgTransform')(ele),
                from_x = from_ele_info.translate[0],
                from_y = from_ele_info.translate[1],
                from_rotate = from_ele_info.rotate,
                to_x = transX,
                to_y = plotH + 20 + bandWidth/2;

            if (this.chart.axes.rotated && yOrient === "bottom") {
                if(yOrient === "bottom"){
                    val = [transX, -this.chart.plotarea.top + bandWidth/2];
                    if (defined(_rot)) {
                        val = [transX, -this.chart.plotarea.top + bandWidth/2];
                    }
                    to_x = val[0];
                    to_y = val[1];
                }
            }
            var to_rotate = _rot,
                x = d3.interpolate(from_x, to_x),
                y = d3.interpolate(from_y, to_y),
                rotate = d3.interpolate(from_rotate, to_rotate);

            return {
                x,
                y,
                rotate
            };
        } else {
            if (this.chart.axes.rotated && yOrient === "bottom") {
                if(yOrient === "bottom"){
                    val = `translate(${[transX, -this.chart.plotarea.top + bandWidth/2]})`;
                    if (defined(_rot)) {
                        val = `translate(${[transX, -this.chart.plotarea.top + bandWidth/2]}) rotate(${_rot})`;
                    }
                    return val;
                }
            }
            var val = `translate(${[transX, plotH + 20 + bandWidth/2]})`;
            if (defined(_rot)) {
                val = `translate(${[transX, plotH + 20 + bandWidth/2]}) rotate(${_rot})`;
            }
            return val;
        }
    };
}
export function racingBoundBox_position() {
    const _this = this,
        KEY = 'racingBoundBox',
        RACE_SUPER_KEY = 'racingLabel',
        SUPER_KEY = 'boundBox';

    _this.helpers.position[KEY] = {};

    _this.helpers.position[KEY].start = {};
    _this.helpers.position[KEY].exit = {};
    _this.helpers.position[KEY].end = {};

    this.helpers.position[KEY].start = (d, i, ele) => {
        return this.helpers.position[RACE_SUPER_KEY].start.transform(d, i, ele);
    };

    this.helpers.position[KEY].end = (d, i, ele) => {
        return this.helpers.position[SUPER_KEY].end.transform(d, i, ele);
    };

    this.helpers.position[KEY].exit = (d, i, ele) => {
        return this.helpers.position[RACE_SUPER_KEY].exit.transform(d, i, ele);
    };
}

export function boundBox_position() {
    const KEY = 'boundBox';
    var _this = this,
        _sereisInfo = _this.subSeries,
        chart = _this.chart,
        properties = _this.properties;

    _this.helpers.position[KEY] = {};

    _this.helpers.position[KEY].start = {};
    _this.helpers.position[KEY].stop_1 = {};
    _this.helpers.position[KEY].end = {};

    _this.helpers.position[KEY].start.transform = (d, i) => {
        var _s = _this.subSeries[i],
            pos = _s.xy || [_s.X, _s.Y],
            rotation = properties.rotation,
            plotH = chart.plotarea.height,
            val = '';

        if (chart.axes.rotated) {
            pos = (_s.xy && [0, _s.xy[1]]) || [0, _s.Y];
        } else {
            pos = (_s.xy && [_s.xy[0], plotH]) || [_s.X, plotH];
        }
        val = `translate(${pos})`;
        if (defined(rotation)) {
            val = `translate(${pos}) rotate(${rotation})`;
        }
        return val;
    };

    _this.helpers.position[KEY].stop_1.transform = (d, i, ele) => {
        var _s = _this.subSeries[i],
            pos = _s.xy || [_s.X, _s.Y],
            val = '',
            eleInfo = $ZC.get('domUtils.getSvgTransform')(ele),
            _rot = eleInfo.rotate;

        if (chart.axes.rotated) {
            var transX = eleInfo.translate[0];
            val = `translate(${[transX, pos[1]]})`;
            if (defined(_rot)) {
                val = `translate(${[transX, pos[1]]}) rotate(${_rot})`;
            }
        } else {
            var transY = eleInfo.translate[1];
            val = `translate(${[pos[0], transY]})`;
            if (defined(_rot)) {
                val = `translate(${[pos[0], transY]}) rotate(${_rot})`;
            }
        }

        return val;
    };

    _this.helpers.position[KEY].end.transform = (d, i) => {
        var _s = _this.subSeries[i],
            pos = _s.xy || [_s.X, _s.Y],
            rotation = properties.rotation,
            val = '',
            _minDist,
            isRotation_pos;

        if (defined(rotation)) {
            (_minDist = 360 - rotation), (isRotation_pos = rotation > 0);
            rotation = _minDist < rotation ? (isRotation_pos ? -_minDist : _minDist) : rotation;
        }
        val = `translate(${pos})`;
        if (defined(rotation)) {
            val = `translate(${pos}) rotate(${rotation})`;
        }

        return val;
    };
}
export function boundBox_style() {
    const KEY = 'boundBox';
    let _this = this,
        { subSeries: _sereisInfo, chart, properties, datalabel_type: labelType, isPolarAxisCategory, isAxisCategory } = _this,
        { dataLabels: config, isMultiLine, chartType } = properties,
        boundboxOptions = getMergedBoundBoxOptions(config, chart),
        hookOptions = getBoundBoxHookConf(boundboxOptions.hook),
        globalDataUpdate = properties.dataUpdate,
        padding = getBoundBoxPadding(config, chart),
        r = getBorderRadius(boundboxOptions.radius),
        strokeWidth = defined(boundboxOptions.strokeWidth) ? +boundboxOptions.strokeWidth : 1,
        strokeDashArray = getDashArray(boundboxOptions.dashStyle, strokeWidth),
        strokeLineCap = getLineCap(boundboxOptions.lineCap, boundboxOptions.dashStyle),
        heightAdjustScale = d3_scaleLinear(),
        plotarea = chart.plotarea,
        trans = properties.trans || [0, 0],
        edge = {
            left: -trans[0] + getFontSizeByHierarchy(config, chart),
            right: plotarea.width - trans[0] - getFontSizeByHierarchy(config, chart)
        };
    _this.helpers.styles[KEY] = {
        display: function(d, i, args) {
            return hasBoundBox(_this.subSeries[i]) ? null : NONE;
        },
        fill: function(d, i, args) {
            var fillColor = getBoundBoxColor(boundboxOptions.fillColor, properties, chart, globalDataUpdate, args.legendOrder);
            return invokeFunction(fillColor, d, getMetaConfigForBoundbox(i, args.legendOrder, properties), chart);
        },
        'fill-opacity': boundboxOptions.fillOpacity,
        stroke: function(d, i, args) {
            var strokeColor = getBoundBoxColor(boundboxOptions.strokeColor, properties, chart, globalDataUpdate, args.legendOrder);
            return invokeFunction(strokeColor, d, getMetaConfigForBoundbox(i, args.legendOrder, properties), chart);
        },
        'stroke-opacity': boundboxOptions.strokeOpacity,
        'stroke-dasharray': strokeDashArray,
        'stroke-linecap': strokeLineCap,
        transition: 'opacity ease .2s',
        opacity: function(d, i, args) {
            if (defined(chart.dataObject.hierarchical) && defined(chart.cache.hierarchical)) {
                var cache = chart.cache.hierarchical;
                if (cache.active) {
                    if (d.active === true) {
                        return 1;
                    } else {
                        return 0.1;
                    }
                }
            }
        },
        strokeWidth: boundboxOptions.strokeWidth,
        d: function(d, i, args) {
            var point = _this.subSeries[i],
                xy = point.xy || [point.X, point.Y];
            // var x = - padding,
            //     y = - padding,
            var x = xy[0] - padding,
                y = xy[1] - padding,
                w = point.width,
                h = point.height,
                dy = point.actualDY !== undefined ? point.actualDY : parseFloat(point.DY) || 0,
                anchor = point.ANCHOR,
                hookPosition = point.hookPosition,
                dyHeight =
                    args.labelType === DATALABELS_LABELTYPE &&
                    arrayIncludes([charttype.funnel, charttype.pyramid], chartType) &&
                    defined(config.y) &&
                    isMultiLine
                        ? properties.txtHgt
                        : h;
            x = anchor === 'middle' ? x - w / 2 : anchor === 'end' ? x - w : x;
            y = y - heightAdjustScale.domain([-0.3, 0.9]).range([dyHeight, 0])(dy);

            if (!isAxisCategory && !isPolarAxisCategory && properties.type !== 'hierarchy') {
                if(isUndefined(_this.subSeries[i].trimmedWidth)){
                    if (x < edge.left) {
                        w -= edge.left - x;
                        x = edge.left;
                    }
                    if (x + w + strokeWidth / 2 + padding > edge.right) {
                        w -= x + w - edge.right + strokeWidth / 2 + padding * 2;
                    }
                }else{
                    if(x < edge.left){
                        x = _this.subSeries[i].X - _this.subSeries[i].trimmedWidth - padding;
                    }
                    w = _this.subSeries[i].trimmedWidth;
                }
                if(hookPosition === RIGHT_ALIGN || hookPosition === LEFT_ALIGN){
                    if(hookPosition === RIGHT_ALIGN || isUndefined(hookPosition)){
                        x = x  - strokeWidth;
                    }
                    w += strokeWidth;
                }
                x += (boundboxOptions.strokeWidth || 1) * (hookPosition === RIGHT_ALIGN ? 0.5 : hookPosition === LEFT_ALIGN ? -0.5 : 0);
            }

            x = x - xy[0];
            y = y - xy[1];
            return getBoundBoxPathDescription(x, y, w, h, r, padding, hookPosition, hookOptions, point.adjustVal);
        },
        id: function(dt, j) {
            return `boundbox${_this.subSeries[j].ID}`;
        },
        levelbycolumn: function(_d, i) {
            let d = _this.actualData ? _this.actualData[_this.actualData.length - 1 - i] : _d.data || _d;
            if (d.levelByColumn) {
                return d.levelByColumn;
            }
        },
        levelbyrow: function(_d, i) {
            let d = _this.actualData ? _this.actualData[_this.actualData.length - 1 - i] : _d.data || _d;
            if (d.levelByRow) {
                return d.levelByRow;
            }
        }
    };
}
