// @zohocharts/zc/src/label/parser/datalabel-parser.js
import { colorscale_helpers } from '../../axis/colorscale-helpers';
import { isCategoryOf } from '../../core/charttype-helpers';
import { charttype } from '../../core/main-zc-init';
import { renderDom } from '../../mark/dom-render';
import { executeMarkOperators } from '../../plot/mark-operator-executor';
import { INNERLABEL_LABELTYPE, PARENTLABELS_LABELTYPE } from '../labeltype-constants';
import { datalabel_encode_collectors } from './datalabel-encode-helpers';
import { ZC_DATALABEL_MARKS_MAPPING } from './datalabel-marks-mapping';
import { datalabel_parser_helpers } from './datalabel-parser-helpers';

export class DatalabelRenderer {
    constructor(args) {
        let {
            datalabel_type,
            ctype,
            chart,
            dataUpdate,
            actualData,
            properties,
            builderC,
            rendererIndex,
            subSeries,
            index,
            subseriesIdx,
            datalabelOptions,
            commonOptions,
           options,
           operator
        } = args;
        this.datalabel_type = datalabel_type;
        this.ctype = ctype;
        this.chart = chart;
        this.dataUpdate = dataUpdate;
        this.actualData = actualData;
        this.conf = chart.rendererConf[rendererIndex];
        this.properties = properties;
        this.rendererIndex = rendererIndex;
        this.builderC = builderC;
        this.subSeries = subSeries;
        this.index = index;
        this.subseriesIdx = subseriesIdx;
        this.datalabelOptions = datalabelOptions;
        this.commonOptions = commonOptions.datalabels;
        this.options = options.datalabels;
        this.markOperator = operator;
        this.animate  = this.commonOptions.animation && this.commonOptions.animation.enabled ||this.options.animation &&this.options.animation.enabled;
        this.isStackedLabel = datalabel_type === 'stacklabel';
        this.isInnerLabel = datalabel_type === INNERLABEL_LABELTYPE || datalabel_type === PARENTLABELS_LABELTYPE;
        this.hasColorScale = colorscale_helpers.isQuantileColorScale(chart)
        this.isPolar = isCategoryOf(chart, properties.chartType, 'polarAxis'),
        this.isHierarchy = isCategoryOf(chart, properties.chartType, 'hierarchy'),
        this.isNonAxisMultiSeriesCategory = isCategoryOf(chart, properties.chartType, 'nonAxisMultiSeries'),
        this.isAxisCategory = isCategoryOf(chart, properties.chartType, 'axis'),
        this.isTreemap = properties.chartType == charttype.treemap,
        this.isPolarAxisCategory = isCategoryOf(chart, properties.chartType, 'polarAxis')
        this.isPackedBubble = properties.chartType == charttype.packedbubble,
        this.isRedraw = this.conf.parser && this.conf.parser.phase || chart.redraw;
        this.setMarkMapping();
    }
    initialize(container) {
        this.container = container;
        datalabel_parser_helpers.call(this);
        this.renderInfo = this.chart.renderer[this.rendererIndex].commonRendererProp().params; //getDatalabelCommonOption(this.conf, this.rendererIndex, this.legendOrder, this.paneid);
        this.scales = {
            x: this.renderInfo.xscale,
            y: this.renderInfo.yscale
        };
        this.collectEncode();
    }
    collectEncode() {
        datalabel_encode_collectors.call(this);
    }
    getAllowedMarkOperator(){
        return ["datalabel"];
    }
    getDefaultMarkParser() {
        let operators = this.getAllowedMarkOperator(),
            markMapping = this.markMapping,
            customOperator = this.markOperator;
        this._mark = executeMarkOperators.call(this, this.data, markMapping, operators, customOperator);
    }
    setMarkMapping(mapping) {
        this.markMapping = (mapping || ZC_DATALABEL_MARKS_MAPPING).call(this);
    }
    render(parent) {
        this._mark.map(d => {
            renderDom(d.mark, parent, this.chart.viewFlow);
        });
    }
    destroy() {
        this.container = null;
    }
}
