// @zohocharts/zc/src/label/parser/datalabel-marks.js
import { isCategoryOf } from '../../core/charttype-helpers';
import { d3_select } from '../../core/d3-constants';
import { charttype } from '../../core/main-zc-init';
import { NONE, NULL, UNDEFINED } from '../../core/zc-constants';
import { Registry } from '../../core/zc-registry';
import { MarksRegistry } from '../../plot/MarksRegistry';
import { getFontFamilyByHierarchy, getFontSizeByHierarchy } from '../../utils/zcdatautils';
import { createElement } from '../../utils/zcsvgutils';
import { bbox, defined, isStylePropertySupported, utils } from '../../utils/zcutils';
import { getBoundBoxPadding, getBoundBoxStrokeWidth, isBoundBoxEnabled } from '../boundBoxUtils';
import { callTrimText } from '../datalabelsUtils';
import { getTspanBound, translateTspan, trimAllTspan } from '../tspan-utils';
import { getBoundBoxParser } from './boundBox-marks';
export function getDefaultParser(data) {
    var _this = this,
        { chart, builderC, properties } = _this,
        enterMarks,
        updateMarks,
        exitMarks,
        data = data ? data : _this.dataUpdate,
        anime_config = builderC.animationOptions,
        { GLT, contentType } = properties,
        children = [],
        racing = this.chart.racing && this.chart.racing.syncTransition,
        contentType = contentType || 'text',
        isShowAs = GLT && GLT.hasShowAsFunction,
        strokeColor = _this.properties.dataLabels.strokeColor,
        applyOutline,
        animate = _this.animate,
        anim_dur = (anime_config && anime_config._duration) || 100, //for other cases other than bar
        anim_delay = (anime_config && anime_config._delay) || 0,
        rotated = chart.axes.rotated;
    
    if (strokeColor === UNDEFINED) {
        strokeColor = chart.systemConf.chart.plot.datalabels.strokeColor; // ZC2362
    }
    applyOutline = strokeColor !== UNDEFINED;
    if (applyOutline && !isStylePropertySupported('paintOrder')) {
        if (!_this.dataLevelStrokeColor) {
            _this.dataLevelStrokeColor = _this.datalabelOptions.stroke;
        }
    }

    var isOrdinal = chart.dataset.getDataType("x",0) === 'ordinal';
    var labelMark = {
        name: 'label',
        selector: '.zc_label',
        mark: isShowAs ? 'g' : 'text',
        data: data,
        dataAccessor:
            racing && isOrdinal
                ? function(_) {
                      if (_this.chart.racing) {
                          d3_select(this).interrupt();
                      }
                      return chart.dataset.getX(_);
                  }
                : null,
        config: {
            stroke: _this.helpers.styles.label.stroke(applyOutline && !isStylePropertySupported('paintOrder') ? false : true),
            strokeWidth: (d, i) => _this.helpers.styles.label['stroke-width'](d, i),
            dy: _this.helpers.position.label.dy
        },
        anime: {},
        properties: labelStyles.call(this),
        onEnd: label_onEnd.call(_this, racing)
    };
    if (racing) {
        var race = chart.racing,
            jumpDuration = race.jumpDuration,
            race_duration = race.duration;
        jumpDuration = jumpDuration > race_duration ? race_duration : jumpDuration;
        let raceLabel_pos = _this.helpers.position.racingLabel;
        enterMarks = {
            transform: (d, k) => {
                return raceLabel_pos.start.transform(d, k);
            },
            x: 0,
            y: 0
        };
        updateMarks = [
            {
                transition: racing,
                transform: {
                    property: 'attrTween',
                    to: function(d, k) {
                        let interpolator = d3.interpolateString(
                            this.getAttribute('transform'),
                            raceLabel_pos.end.transform(d, k, this)
                        );
                        return function(t) {
                            return interpolator(t);
                        };
                    }
                },
                x: 0,
                y: 0
            }
        ];
        exitMarks = {
            transition: racing,
            transform: function(d, k) {
                return raceLabel_pos.exit.transform(d, k, this);
            }
        };
        labelMark.config.enter = enterMarks;
        labelMark.anime.update = updateMarks;
        labelMark.anime.exit = exitMarks;
    } else if (animate) {
        var label_pos = _this.helpers.position.label;
        enterMarks = {
            transform: (d, i) => {
                return label_pos.start.transform(d, i);
            },
            x: (d, i) => {
                return label_pos.start.x(d, i);
            },
            y: (d, i) => {
                return label_pos.start.y(d, i);
            }
        };
        updateMarks = [
            {
                duration: _this.isRedraw == 'redraw' ? anim_dur : 0,
                delay: anim_delay,
                transform: {
                    property: 'attrTween',
                    to: function(d, i) {
                        var interpolator = d3.interpolateString(this.getAttribute('transform'), label_pos.stop_1.transform(d, i, this));
                        //if from is rotate(0,10,10) nd to is translate(0,0) - case not handled
                        //FEATURESEVENTS_SCROLL_2 - apply renderOnInit with animate enabled true, - chnage from bar to line - from will be rotate(0,10,10) but to will be null
                        return function(t) {
                            return interpolator(t);
                        };
                    }
                },
                x: function(d, i) {
                    return label_pos.stop_1.x(d, i, this);
                },
                y: function(d, i) {
                    return label_pos.stop_1.y(d, i, this);
                }
            },
            {
                duration: anim_dur,
                delay: anim_delay,
                transform: {
                    property: 'attrTween',
                    to: function(d, i) {
                        let interpolator = d3.interpolateString(this.getAttribute('transform'), label_pos.end.transform(d, i));
                        return function(t) {
                            return interpolator(t);
                        };
                    }
                },
                x: (d, i) => {
                    return label_pos.end.x(d, i);
                },
                y: (d, i) => {
                    return label_pos.end.y(d, i);
                }
            }
        ];
        exitMarks = {
            duration: 0
        };
        labelMark.config.enter = enterMarks;
        labelMark.anime.update = updateMarks;
        labelMark.anime.exit = exitMarks;
    } else {
        var label_pos = _this.helpers.position.label;
        updateMarks = {
            transform: (d, i) => {
                return label_pos.end.transform(d, i);
            },
            x: (d, i) => {
                return label_pos.end.x(d, i);
            },
            y: (d, i) => {
                return label_pos.end.y(d, i);
            },
            translateX: (d, i) => {
                let _translate = label_pos.end.translateXY(d, i);
                return _translate ? _translate[0] : null;
            },
            translateY: (d, i) => {
                let _translate = label_pos.end.translateXY(d, i);
                return _translate ? _translate[1] : null;
            }
        };
        if(chart.dataObject.sankeyLayout && chart.dataObject.sankeyLayout.zoomTransform){
            updateMarks.translateX =  (d, i) => {
                let _translate = label_pos.end.translateXY(d, i);
                return _translate ? _translate[0] : null;
            }
            updateMarks.translateY = (d, i) => {
                let _translate = label_pos.end.translateXY(d, i);
                return _translate ? _translate[1] : null;
            }
        }
        exitMarks = {
            duration: 0
        };
        labelMark.config.update = updateMarks;
        labelMark.anime.exit = exitMarks;
        //we store it here since, for none animation case, incase of trim we will need exact x and y values to calculate bbox
        //if added in anime - only at the end x and y updated, so bbox calculation was not proper.
    }

    labelMark.config[contentType] = (d, i) => _this.helpers.text.label(d, i);
    

    if (applyOutline && !isStylePropertySupported('paintOrder')) {
        _this.isOutLineGroup = true;
        var outLine = {
            name: 'outline',
            selector: '.outline',
            mark: 'text',
            data: data,
            config: {
                update: updateMarks,
                exit: exitMarks,
                stroke: _this.helpers.styles.label.stroke(true),
                strokeWidth: (d, i) => _this.helpers.styles.label['stroke-width'](d, i, true)
            },
            properties: labelStyles.call(this, true),
            onEnd: label_onEnd.call(_this, racing)
        };
        outLine.config[contentType] = (d, i) => _this.helpers.text.label(d, i);
        children.push({
            name: 'outputGroup',
            selector: '.outputGroup',
            mark: 'g',
            data: [1],
            children: [outLine],
            config: {
                strokeWidth: "inherit"
            }
        });
    }
    children.push(labelMark);
    var boundBoxMark = getBoundBoxParser.call(_this, data, animate, anime_config);
    boundBoxMark ? children.push(boundBoxMark) : null;
    
    return {
        mark: 'g',
        data: [1],
        selector: `.subseries_${_this.subseriesIdx}`,
        properties: {
            attrs: {
                class: (d, i) => {
                    return `subseries_${_this.subseriesIdx}`;
                },
                transform: this.subSeries.translation
            }
        },
        children: children,
        onEnd: function(ele) {
            ele.style('stroke-width', 'inherit');
        }
    };
}

MarksRegistry.add('datalabel.default', getDefaultParser);

//helper functions
function labelStyles(isOutLineText) {
    var _this = this,
        dataLabels = _this.properties.dataLabels,
        clip = dataLabels.textOverlap === 'clip';
    return {
        styles: {
            display: (d, i) => _this.helpers.styles.label.display(d, i),
            fill: (d, i) => _this.helpers.styles.label.fill(d, i),
            'pointer-events': NONE,
            opacity: (d, i) => _this.helpers.styles.label.opacity(d, i, isOutLineText)
        },
        attrs: {
            dy: (d, i) => _this.helpers.styles.label.dy(d, i),
            'text-anchor': (d, i) => _this.helpers.styles.label['text-anchor'](d, i),
            id: (d, i) => _this.helpers.styles.label.id(d, i),
            'clip-path': !clip ? null : (d, i) => _this.helpers.styles.label['clip-path'](d, i),
            levelbycolumn: (d, i) => _this.helpers.styles.label.levelbycolumn(d, i),
            levelbyrow: (d, i) => _this.helpers.styles.label.levelbyrow(d, i),
            'filtered-opacity': (d, i) => _this.helpers.styles.label['filtered-opacity'](d, i)
        }
    };
}

function label_onEnd(racing) {
    var _this = this,
        { ctype: chartType, isPolar, isAxisCategory, isPackedBubble, isTreemap, isStackedLabels, isInnerLabel, properties, chart } = _this,
        { GLT, contentType, dataLabels } = properties,
        fontSize = getFontSizeByHierarchy(dataLabels, chart),
        boundBoxPadding = isBoundBoxEnabled(dataLabels, chart)
            ? getBoundBoxPadding(dataLabels, chart) + getBoundBoxStrokeWidth(dataLabels, chart)
            : 0,
        hasLevelMarker = chartType == charttype.bullet || chartType == charttype.dial,
        animate = _this.animate,
        isMultiLineAnimation = properties.isMultiLine && animate,
        fontFam = getFontFamilyByHierarchy(dataLabels, chart);

    return _end;
    function _end(d) {
        d.nodes().map((d, i) => {
            if(racing){
                //in racing we will give transform-translate, not x,y so if 'x' is there that element will get positioned twice.
                d3_select(d).selectAll(".zcTspan").attr("x", null);
            }
            if (!isMultiLineAnimation && racing && Number(d.getAttribute('x')) != 0 && Number(d.getAttribute('y')) != 0) {
                /*
                 * if isMultiLineAnimation - the values of transform,x,y will already be as below
                 * racing - initial transform,x,y will not be applied as per racing since syncTransition is avaialable after onload
                 * so inorder to animate we modify the transform,x nd y values
                 * we need to do this only once
                 */
                var details = $ZC.get('domUtils.getSvgTransform')(d);
                d.setAttribute('transform', `translate(${d.getAttribute('x')},${d.getAttribute('y')}) rotate(${details.rotate})`);
                d.setAttribute('x', 0);
                d.setAttribute('y', 0);
            }

            var k = i;
            if (properties.isMultiLine) {
                for (var p = 0; p < d.childNodes.length; p++) {
                    if (animate || racing) {
                        // for cases in which animation enabled and racing cases
                        d.childNodes[p].setAttribute('x', 0);
                    } else {
                        d.childNodes[p].setAttribute('x', _this.subSeries[k].X);
                    }

                    // if(!Number(d.childNodes[p].getAttribute('x'))){
                    //     d.childNodes[p].setAttribute('x', d.getAttribute('x'));
                    // }
                }
            }
            if (
                !isAxisCategory &&
                !isTreemap &&
                !isPolar &&
                !hasLevelMarker &&
                !isPackedBubble &&
                !isStackedLabels &&
                !isInnerLabel &&
                !chart.isGeoMap &&
                !(GLT && GLT.hasShowAsFunction) // ZC-T2092
            ) {
                if (contentType === 'html') {
                    if (_this.subSeries[k].display !== false) {
                        let useBBox = dataLabels.preCalculation && dataLabels.preCalculation.useBBox,
                            newX = null;
                        var ele = d3_select(d);

                        if (/*_this.properties.isMultiLine && */ animate) {
                            var rotation = _this.properties.rotation,
                                _sereisInfo = _this.subSeries,
                                _s = _sereisInfo[k],
                                pos = _s.xy || [_s.X, _s.Y];
                            ele.attr('x', _s.X)
                                .attr('y', _s.Y)
                                .attr('transform', rotation ? `rotate(${rotation} ${pos})` : NULL);
                            ele.selectAll('.zcTspan').attr('x', _s.X);
                        }
                        var tspanInfo = getTspanBound(ele, fontSize, useBBox);
                        trimAllTspan(
                            ele,
                            chart,
                            properties.trans,
                            properties.txtHgt,
                            null,
                            tspanInfo,
                            _this.subSeries[k].isLeft ? boundBoxPadding : 0,
                            undefined,
                            dataLabels.preCalculation
                        );

                        if (/*_this.properties.isMultiLine && */ animate) {
                            var rotation = _this.properties.rotation,
                                _sereisInfo = _this.subSeries,
                                _s = _sereisInfo[k],
                                pos = _s.xy || [_s.X, _s.Y];
                            ele.attr('x', 0)
                                .attr('y', 0)
                                .attr('transform', rotation ? `translate(${pos}) rotate(${rotation})` : `translate(${pos})`);
                            ele.selectAll('.zcTspan').attr('x', 0);
                        }
                        if (_this.subSeries[k].ANCHOR === 'middle') {
                            var TX = (_this.subSeries[k].width - bbox(d).width) / 2;
                            var flip = _this.subSeries[k].isLeft ? 1 : -1;
                            newX = _this.subSeries[k].X + TX * flip;
                        }

                        if (defined(newX)) {
                            ele.attr('x', newX);
                            translateTspan(ele, newX);
                        }
                    }
                } else {
                    callTrimText(d3_select(d), chart, properties.trans, properties.txtHgt);
                    if(d.textContent === '..'){
                        d.textContent = '';
                        _this.subSeries[k].display = false;
                    }
                    var trimmedWidth = utils.fontUtils.getTextWidth(d.textContent, fontSize, fontFam);
                    _this.subSeries[k].trimmedWidth = trimmedWidth;
                }
            }
        });

        //when extended is added for mark
        if (Registry.getComponent('defaultDatalabelParser')()) {
            Registry.setComponent('datalabel.default', Registry.getComponent('defaultDatalabelParser')());
        }
    }
}
