// @zohocharts/zc/src/axis/radial/parser/marks/radial-ticks-marks.js
import { ParserRegistry } from "../../../../core/parser-registry";
import { deleteProperties, mergeJSON } from "../../../../utils/zcutils";
import { _cartesianTicks as ticks } from "../../../cartesian/parser/marks/cartesian-ticks-marks";

export function _radialTicks(data) {
    const self = this, KEY = 'ticks';

    const { helpers } = this,
        _position = helpers.position[KEY],
        _styles = helpers.styles[KEY];

    let ticksMarks = ticks.call(this, data);
    let minorTicksMarks = minorTicks.call(this, data, $selector(ticksMarks.children, '.minorTicks'));
    let majorTicksMarks = majorTicks.call(this, data, $selector(ticksMarks.children, '.majorTicks'));
    let tickLabelMarks = tickLabel.call(this, data, $selector(ticksMarks.children, '.ticklabel'))

    //* remove unwanted properties here
    deleteProperties(ticksMarks, ['config', 'anime']);

    //* customize parser values here
    const customTicksMarks = {
        type: 'html',
        properties: {
            attrs: {
                transform: _position.end,
                opacity: 1
            },
            styles: {
                // display: //* ln 636 key='tick.style'
            }
        }
    };

    //* merge custom with default value;
    mergeJSON(ticksMarks, customTicksMarks, true, true);
    return ticksMarks;
}
ParserRegistry.axis('radial.ticks', _radialTicks);

function minorTicks(data, parser) {
    const KEY = 'tickmark';
    const { helpers } = this,
        { position, styles } = helpers,
        _position = position[KEY],
        _styles = styles[KEY];

    delete parser.config;

    const marks = {
        type: 'html',
        styles: {
            'stroke-width': _styles.strokeWidth
        }
    }

    mergeJSON(parser, marks, true, true);
    return parser;
}

function majorTicks(data, parser) {
    const KEY = 'gridline';
    const { helpers, component, renderer } = this,
        { position, styles } = helpers,
        { renderGrid, isGridShow } = component,
        // gridSize = renderer.majorTicks(),
        _position = position[KEY],
        _styles = styles[KEY];

    deleteProperties(parser, ['config', 'properties']);

    const marks = {
        type: 'html',
        mark: 'path',
        // data: isGridShow && renderGrid ? [d] : [],
        config: {
            strokeWidth: _styles.strokeWidth
        },
        properties: {
            attrs: {
                transform: _position.transform,
                d: _position.d
            },
            styles: {
                stroke: _styles.stroke,
                // 'stroke-width': _styles.strokeWidth,
                fill: _styles.fill
            }
        }
    }

    mergeJSON(parser, marks, true, true);
    return parser;

}

function tickLabel(data, parser) {
    const KEY = 'ticklabel';
    const { helpers } = this,
        _position = helpers.position[KEY],
        _styles = helpers.styles[KEY];;

    const marks = {
        type: 'html',
        properties: {
            attrs: {
                x: _position.x,
                y: _position.y,
                dy: _position.dy,
                transform: _position.transform
            },
            styles: {
                'font-size': _styles.fontSize,
                'font-family': _styles.fontFamily,
                'text-anchor': _styles.textAnchor
            }
        }
    }

    deleteProperties(parser, ['config']);
    mergeJSON(parser, marks, true, true);
    return parser;
}

function $selector(parser, selector) {
    return parser.find(p => p.selector == selector);
}
