// @zohocharts/zc/src/scale-parser/legend/color/color-domain-parser.js
import { d3_values, pick } from '../../../utils/zcutils';
import { fetchForColorDomainParser } from '../../fetch-operators';
import { d3_ascending } from '../../../core/d3-constants';
import { filterArrayWithInRange } from '../../../scale/operators/numeric/filter-by-range';
import { getLegendRanges } from '../../../axis/colorscale-helpers';
import { ScaleRegistry } from '../../../scale/scale-registry';
import { ALLOWED_OPERATORS } from '../../../scale/operators/allowed-operators';

export function colorDomainParser(encodeType, index, config, chart) {
    let band = config.colorBand,
        stops = band.stops,
        ranges = band.ranges,
        id = encodeType + pick(index, ''),
        processor = chart.dataObject,
        datatype = chart.dataset.getDataType(encodeType);

    let domainOperators = {},
        defaultOperators = fetchForColorDomainParser(datatype);
    /*
        EXTREMES
     */
    if (defaultOperators.has('extremes')) {
        domainOperators.extremes = [
            {
                _type: 'extremesByData',
                data: processor.processedseriesData,
                indices: [chart.dataset.clridx]
            }
        ];
    }

    /*
        Custom min max range
     */
    domainOperators.min = band.minRange;
    domainOperators.max = band.maxRange;

    if (ranges && ranges.length) {
        /* Ranges*/
        domainOperators.ranges = true;
        domainOperators.sortedRanges = {
            datatype: datatype,
            config: band
        };

        //pair and splicemax
    } else if (stops && stops.length) {
        /*Stops*/
        domainOperators.stops = true;
        domainOperators.filteredStops = {
            array: d3_values(stops).sort(d3_ascending),
            min: band.minRange,
            max: band.maxRange
        };
        //spliceminmax
    }
    return domainOperators;
}

export function sortRanges(operatorC) {
    return getLegendRanges(operatorC.config, operatorC.datatype, true, operatorC.clone);
}

export function filterStops(operatorC) {
    return filterArrayWithInRange(operatorC.array, operatorC.min, operatorC.max);
}
ScaleRegistry.add('color.domainParser', colorDomainParser);

ScaleRegistry.add('color.sortRanges', sortRanges);
ScaleRegistry.add('color.filterStops', filterStops);

ALLOWED_OPERATORS.set('filteredStops', ScaleRegistry.get('color.filterStops'));
ALLOWED_OPERATORS.set('sortedRanges', ScaleRegistry.get('color.sortRanges'));
