// @zohocharts/zc/src/scale-parser/legend/color/color-range-parser.js
import { getAxisOrient } from '../../../axis/axisutils';
import { ORDINAL_DATATYPE } from '../../../core/datatype-constants';
import { getLegendType } from '../../../axis/colorscale-helpers';
import { pick } from "../../../utils/zcutils";
import { fetchForColorRangeParser} from "../../fetch-operators";
import {ScaleRegistry} from "../../../scale/scale-registry";

export function colorRangeParser(encodeType, index, config, chart) {
    let band = config.colorBand,
        stops = band.stops,
        ranges = band.ranges,
        id = encodeType + pick(index, ''),
        processor = chart.dataObject,
        datatype = chart.dataset.getDataType(encodeType),
        isOrdinal = datatype == ORDINAL_DATATYPE,
        orient = getAxisOrient(encodeType, chart.axes.rotated),
        legendType = getLegendType(config, chart),
        isContinousLegend = legendType !== 'discrete';

    let rangeOperators = {},
        defaultOperators = fetchForColorRangeParser(datatype);

    if (defaultOperators.has('colorRanges')) {
        rangeOperators.colorRanges = {
            colors: chart.imagePallete || chart.seriesColor
        };
    }

    if (ranges && ranges.length) {
        rangeOperators.sortedRanges = {
            datatype: datatype,
            config: band
        };
        rangeOperators.mapArrayRanges = {
            ranges: ranges,
            isOrdinal: isOrdinal,
            isContinousLegend: isContinousLegend
        };
    } else if (!isOrdinal && stops && stops.length) {
        rangeOperators.mapDistinctRanges = {
            stops: stops
        };
    }
    return rangeOperators;
}

ScaleRegistry.add('color.rangeParser', colorRangeParser);
