// @zohocharts/zc/src/area/areaRangeOptions.js
import { defaultTheme as options } from '../core/defaulttheme';

options.arearange = function() {
    return {
        strokeOpacity: 0.5,
        outerPadding: 0.02,
        fillOpacity: 0.4,
        mode: 'linear', //linear,stepBefore,stepAfter,cardinal
        stacked: {
            enabled: false,
            showAs: 'values' //values,percent
        },
        marker: {
            enabled: false,
            outerFillColor: 'transparent',
            outerStrokeOpacity: 0.5,
            size: 1 //1,2....
        },
        hoveroptions: {
            innerStrokeWidth: 2,
            outerFillOpacity: 0.5,
            outerStrokeOpacity: 0.1
        },
        datalabels: {
            showAs: 'y' //x,y,percent
        },
        gradients: {
            // linear || none
            options: {
                linear: {
                    x2: 50
                }
            }
        },
        animation: {
            type: 'vertPathTweenAll', //horiPathTween || vertPathTween || vertPathTweenAll || pointTween
            marquee: {
                enabled: false,
                duration: 300
            },
            blink: {
                enabled: false,
                type: 'twinkle', //twinkle | fade
                duration: 800
            }
        }
    };
};

export var _arearangeOptions = options.arearange;
