// @zohocharts/zc/src/area/area-highlight.js
import { _super, inherit, defined } from '../utils/zcutils';
import { ScatterHighlighter } from '../scatter/scatterHighlight';
import { getPolygonPoints } from './polyPoints';
import { d3_merge, d3_min, d3_max } from '../core/d3-constants';
import { NULL } from '../core/zc-constants';
import { patternFillEffect } from '../events/visualEffects/highlight/pattern-fill';
import { appendEle } from '../utils/zcsvgutils';
import { effects } from '../effects/zceffects';
import { HIGHLIGHT_CONTEXT } from '../events/visualEffects/highlight/highlight-context';
import { isPointVisible } from '../events/events-utils';

export function AreaHighlighter(chart, renderer) {
    this.chart = chart;
    this.renderer = renderer;
    return this;

    // extend(AreaHighlighter.prototype, LineHighlighter().prototype);
}
inherit(AreaHighlighter, ScatterHighlighter);
var proto = AreaHighlighter.prototype;

proto._defaultH = function() {
    let defaultH = [
        {
            shape: 'marker',
            path: _super(this).path,
            data: _super(this).data,
            context: _super(this).renderingContext,
            translate: _super(this).translateContext
        },
        {
            shape: 'poly',
            path: this.path,
            data: this.data,
            context: this.renderingContext,
            translate: _super(this).translateContext,
            styles: point => {
                return patternFillEffect.call(this, point, 'inherit');
            }
        }
    ];
    if (this.renderer.commonRendererProp().params.isRangeChart) {
        defaultH.push({
            shape: 'marker',
            path: _super(this).path,
            data: _super(this).data,
            context: _super(this).renderingContext,
            translate: _super(this).translateContext,
            type: 'end'
        });
    }
    return defaultH;
};

proto.data = function(point) {
    var rp = this.renderer.commonRendererProp().params;
    var chart = this.chart,
        renderer = this.renderer,
        axisrotated = rp.axisrotated,
        stacked = rp.stacked,
        layout = renderer.layout[point.subSeriesIndex],
        x = axisrotated ? layout.x1() : layout.x(),
        y = axisrotated ? layout.y() : layout.y1(),
        x0 = axisrotated ? layout.x0() : layout.x(),
        y0 = axisrotated ? layout.y() : layout.y0(),
        isDefined = layout.defined();
    point._polydata = this.poly(point);

    return function() {
        if (!chart.cache.data[point._key]) {
            chart.cache.data[point._key] = {};
        }
        var cache = chart.cache.data[point._key],
            itemIndex = defined(point.counter) ? point.counter : point.itemIndex;

        cache._x = point._x = cache._x || x(point, itemIndex);
        cache._y = point._y = cache._y || y(point, itemIndex);
        cache._x0 = point._x0 = cache._x0 || x0(point, itemIndex);
        cache._y0 = point._y0 = cache._y0 || y0(point, itemIndex);
        cache._uid = point._uid = chart.dataset.getX(point);
        cache._defined = point._defined = defined(cache._defined) ? cache._defined : isDefined(point);
        cache._size = point._size = cache._size || renderer.getSymbolSize && renderer.getSymbolSize(point);
        cache._visible = point._visible = isPointVisible(chart, point) && (stacked ? !point.isDummy : true);
        cache._point = point;
        return point;
    };
};

proto.poly = function(point) {
    var chart = this.chart,
        renderer = this.renderer;
    return function() {
        if (!chart.cache.data[point._key]) {
            chart.cache.data[point._key] = {};
        }
        var cache = chart.cache.data[point._key],
            flatten = renderer.seriesdataAfterUpdate[point.subSeriesIndex];

        cache._poly = point._poly = cache._poly || getPolygonPoints(flatten, point);
        return point._poly;
    };
};

proto.options = function(point) {
    var rp = this.renderer.commonRendererProp().params,
        commonOptions = _super(this).commonOptions.call(this, point),
        symbol = commonOptions.symbol,
        size = commonOptions._size || rp.gvp('strokeWidth'),
        interpolation = rp.gvp('mode'),
        fillopacity = rp.hoverOptions.fillOpacity || 0.2;

    commonOptions.size = {
        outer: size * 0.9,
        inner: size * 0.4
    };
    commonOptions.fillopacity = fillopacity;
    commonOptions.interpolation = interpolation;

    return commonOptions;
};

proto.renderingContext = function() {
    return HIGHLIGHT_CONTEXT.get('area').call(this, this.chart);
};

proto.path = function(point, model) {
    point = point._data();
    var options = this.options(point),
        poly = this.poly(point)(),
        axisrotated = this.chart.axes.rotated,
        context = model.context.call(this),
        pathContext,
        rp = this.renderer.commonRendererProp().params,
        area = this.renderer.layout[point.subSeriesIndex];
    area.context(NULL);
    context.style('display', '').attr('transform', `translate(${this.chart.plotarea.left},${this.chart.plotarea.top})`);

    if (options.interpolation == 'linear') {
        var polyPoints = d3_merge(poly.points);
        pathContext = appendEle(context, 'polygon', [poly], '.', 'class', `highlight${point.paneid}${point.seriesIndex}`);
        pathContext.attr('points', polyPoints.toString()).styles({
            'fill-opacity': options.fillopacity || 0.2,
            display: ''
        });
    } else {
        var points = poly.points;
        var x0 = d3_min(points, d => d[0]),
            x1 = d3_max(points, d => d[0]),
            y0 = d3_min(points, d => d[1]),
            y1 = d3_max(points, d => d[1]);

        var x = x0,
            y = 0,
            w = x1 - x0,
            h = this.chart.plotarea.height;
        if (rp.axisrotated) {
            (x = 0), (y = y0), (w = this.chart.plotarea.width), (h = y1 - y0);
        }
        var properties = {
                name: 'rect',
                id: this.chart.id,
                x: x + ((axisrotated ? 0 : this.chart.scrolledX) || 0),
                y: y + ((axisrotated ? 0 : this.chart.scrolledY) || 0),
                width: w,
                height: h
            },
            args = {
                chartObj: this.chart
            };
        var clipID = effects.get(effects.svg_clips, 'rect', properties, args);

        pathContext = appendEle(context, 'path', [1], '.', 'class', `highlight${point.paneid}${point.seriesIndex}`);
        pathContext
            .datum(this.renderer.seriesdataAfterUpdate[point.subSeriesIndex])
            .attr('d', area)
            .attr('stroke', '')
            .attr('clip-path', `url(${this.chart.defsLocation}#${clipID})`)
            .styles({
                'fill-opacity': options.fillopacity,
                display: ''
            });
    }
    return pathContext;
};
