// @zohocharts/zc/src/area/areachart.js
//$Id$
//areachart.js
import { svg_plot_renderer } from '../core/main-zc-init';
import { DataProcessor } from '../core/data-processor';
import { d3_values, defined, mergeJSON, pInt, utils } from '../utils/zcutils';
import { getAreaLayout } from './area-layout';
import { gradient_helpers } from '../effects/gradient-helpers';
import { colorscale_helpers } from '../axis/colorscale-helpers';
import { getAnimationConf, interTween } from '../utils/animationutils';
import { effects } from '../effects/zceffects';
import { mathMax, mathMin, NULL } from '../core/zc-constants';
import { getFillColor } from '../utils/colorutils';
import { d3_interpolate, d3_select } from '../core/d3-constants';
import { ORDINAL_DATATYPE } from '../core/datatype-constants';
import { appendEle } from '../utils/zcsvgutils';
import {
    HORIPATHTWEEN_ANIMATION,
    POINTBYPOINT_ANIMATION,
    POINTTWEEN_ANIMATION,
    STROKETWEEN_ANIMATION,
    VERTPATHTWEEN_ANIMATION,
    VERTPATHTWEENALL_ANIMATION
} from '../core/animation-constants';
import { AreaHighlighter } from './area-highlight';
import { WebHighlighter } from './web-highlight';
import { getChartTypeByIndex } from '../core/charttype-helpers';
import { setOpacityForTickFilterElements } from '../plot/svg/set-pivot-attr';
import { isFacet } from '../axis/hierarchy-axis-utils';

/**
 *
 * @returns {chart} area and arearange renderer.
 */
svg_plot_renderer.area = function() {
    var commonRendererProp,
        xscale,
        yscale,
        dataset,
        yaxiscolumnorder,
        index = 0,
        stacked = false,
        axisrotated = false,
        redraw = false,
        line = svg_plot_renderer.line?.(),
        baseLine =  svg_plot_renderer.line?.();
    var area = [],
        color,
        labelQueue,
        commonPlotOptions,
        systemconf,
        hasNegValue,
        chartObj,
        transitionDuration;
    var shiftInfo,
        shiftDiff,
        dataProcessorHelpers = DataProcessor.helpers;

    function chart(selection) {
        selection.each(function(data) {
            //renderer config values
            var rp = commonRendererProp.params;
            (chartObj = rp.chartObj),
                (xscale = rp.xscale),
                (yscale = rp.yscale),
                (stacked = rp.stacked),
                (index = rp.datasetIndex),
                (axisrotated = rp.axisrotated),
                (commonPlotOptions = rp.commonPlotOptions),
                (color = rp.color),
                (labelQueue = rp.labelQueue),
                (hasNegValue = rp.hasNegValue),
                (yaxiscolumnorder = rp.yaxiscolumnorder),
                (dataset = rp.dataset),
                (systemconf = chartObj.systemConf);
            var renderingOrder = rp.renderingOrderIndex,
                legendOrder = rp.legendOrderIndex,
                rendererIndex = rp.rendererIndex;

            var defsLocation = chartObj.defsLocation,
                isPolar = chartObj.dataObject.isPolarAxisCategory;

            var parser_data, data1;
            if (chartObj.viewFlow && chartObj.viewFlow.marks._default.chart.length) {
                parser_data = data[index];
                data1 = parser_data.data;
            }

            var currentdata = data1 || data[index];
            var data_bind = { data: currentdata, index: renderingOrder };
            stacked? data_bind.stackIndex = ((data.length * 0) + renderingOrder) : null;
            var plotoptions = currentdata.plotoptions;
            var currentseriesdata = currentdata.data;
            var paneid = isFacet(chartObj) ? `${currentdata.row}_${currentdata.column}` : '0_0';
            var cType_Obj = getChartTypeByIndex(chartObj, paneid, index),
                cType = cType_Obj.name;

            var gvp = (commonRendererProp.params.gvp = utils.getValueOfPlotOptions(plotoptions, commonPlotOptions));
            var fillOpacity = gvp('fillOpacity'),
                strokeWidth = gvp('strokeWidth'),
                strokeColor = gvp('strokeColor'),
                hasColorScale = colorscale_helpers.isQuantileColorScale(chartObj),
                markeroptions = (commonRendererProp.params.markerOptions =
                    plotoptions && plotoptions.marker
                        ? mergeJSON(commonPlotOptions.marker, plotoptions.marker, true)
                        : commonPlotOptions.marker),
                hoverOptions = (commonRendererProp.params.hoverOptions =
                    plotoptions && plotoptions.hoveroptions
                        ? mergeJSON(commonPlotOptions.hoveroptions, plotoptions.hoveroptions, true)
                        : commonPlotOptions.hoveroptions),
                markerEnabled = hasColorScale || markeroptions.enabled;
            var borderEnabled =
                (strokeWidth != null && strokeWidth != '' && pInt(strokeWidth) != 0) ||
                (strokeColor != null && strokeColor != '') ||
                markerEnabled;

            //Animation variables
            var animationOptions = getAnimationConf(commonPlotOptions, plotoptions, chartObj, cType);
            var isAnimationEnable = animationOptions.enabled,
                easingType = animationOptions.easingType,
                animationType = animationOptions.type;
            transitionDuration = animationOptions.duration;
            animationType = animationType == VERTPATHTWEENALL_ANIMATION ? VERTPATHTWEEN_ANIMATION : animationType;

            var datatypeX = chartObj.dataset.getDataType('x'),
                datatypeY = chartObj.dataset.getDataType('y', yaxiscolumnorder),
                xCategories = datatypeX == ORDINAL_DATATYPE ? xscale.domain() : null,
                yCategories = datatypeY == ORDINAL_DATATYPE ? yscale.domain() : null,
                categories =
                    datatypeX == ORDINAL_DATATYPE && !defined(chartObj.dataObject.hierarchical)
                        ? xCategories
                        : chartObj.dataObject.getCategories('x');

            var yReversed = systemconf.chart.axes.yaxis[yaxiscolumnorder].reversed;
            if (datatypeX != ORDINAL_DATATYPE && categories.length == 0 && isAnimationEnable && animationType == POINTTWEEN_ANIMATION) {
                categories = chartObj.dataObject.findXCategories();
            }

            var isRangeChart = (rp.isRangeChart = cType == 'arearange');
            var yrangePos = isRangeChart ? 1 : null;

            chart.seriesdataAfterUpdate = [];
            chart.layout = [];
            currentseriesdata.forEach(function(d, i) {
                var serData;
                if (datatypeX == ORDINAL_DATATYPE || datatypeY == ORDINAL_DATATYPE) {
                    serData = chartObj.dataObject.updateSeriesdatabyCategories(d3_values(d), xCategories, yCategories, yaxiscolumnorder);
                } else {
                    serData = d;
                }

                if (chartObj.renderVisiblePointsOnly || shiftInfo != null) {
                    //for pushdatapoint allow filter for highlight
                    serData = DataProcessor.helpers.filterPointsInViewport(chartObj, serData, cType, 'x', index);
                }

                chart.seriesdataAfterUpdate[i] = serData;
                if (chartObj.systemConf.chart.axes.xaxis.reversed) {
                    serData = serData.slice(0).reverse();
                }
                shiftDiff =
                    shiftInfo != null
                        ? shiftDiff == null || shiftDiff == 0
                            ? chartObj.dataObject.getXYvalue(
                                  axisrotated,
                                  stacked,
                                  xscale,
                                  yscale,
                                  yaxiscolumnorder,
                                  serData[serData.length - 1],
                                  'x'
                              ) -
                              chartObj.dataObject.getXYvalue(
                                  axisrotated,
                                  stacked,
                                  xscale,
                                  yscale,
                                  yaxiscolumnorder,
                                  serData[serData.length - 2],
                                  'x'
                              )
                            : shiftDiff
                        : 0;

                area[i] = getAreaLayout(rp, shiftDiff, shiftInfo, serData, i, paneid);
                chart.layout[i] = area[i];
            });

            var container = d3_select(this);
            chart.update = function() {
                container.transition().call(chart);
            };
            var dynamicPosition = rp.dynamicSeriesPosition;
            if (dynamicPosition === undefined) {
                dynamicPosition = index;
            }
            var series;
            if (isPolar) {
                index = dynamicPosition;
                series = container.selectAll(`g.areaseries_${index}`).data([data_bind]);
                series
                    .enter()
                    .insert('g', ':first-child')
                    .merge(series)
                    .attr('class', `areaseries_${index}`);
                series = container.selectAll(`g.areaseries_${index}`);
            } else {
                let insertBefore;
                if (stacked) {
                    // For ordering series to fix overlapping of markers
                    let areaseries_n = this.querySelector(`g.areaseries_${dynamicPosition-1}`);
                    insertBefore = areaseries_n ? `g.${areaseries_n.className.baseVal} + *` : null;
                }
                series = appendEle(container, 'g', [data_bind], '.', 'class', `areaseries_${dynamicPosition}`, undefined, undefined, {insertBefore : insertBefore});
            }

            if (borderEnabled || markerEnabled) {
                line.commonRendererProp(commonRendererProp).shiftInfo(shiftInfo);
                line.chartRendered = false;

                if (isRangeChart) {
                    line.yrangePos(1);
                    baseLine.commonRendererProp(commonRendererProp)     // moved baseLine var to global to avoid resetting redraw value
                        .shiftInfo(shiftInfo)
                        .yrangePos(0);
                    baseLine.chartRendered = false;
                    container.call(baseLine);
                }
                container.call(line);
            } else {
                container.selectAll(`g.lineseries_${dynamicPosition},g.scatterseries_${dynamicPosition}`).remove();
            }

            currentseriesdata.forEach(function(d, i) {
                var serData = chart.seriesdataAfterUpdate[i];
                var baseData = serData;
                var basedataShift,
                    ShiftedData,
                    dummyDataLength = 0;

                if (isAnimationEnable) {
                    if (animationType == POINTTWEEN_ANIMATION) {
                        baseData = serData.map(function(d, i) {
                            return Object.create(serData[0]);
                        });
                    } else {
                        var idx = animationType == HORIPATHTWEEN_ANIMATION ? dataset.xidx : dataset.yidx(yaxiscolumnorder);
                        var scaleObj = animationType == HORIPATHTWEEN_ANIMATION ? xscale : yscale;
                        var baseIdx = isPolar && yReversed && !stacked ? scaleObj.domain().length - 1 : 0;
                        baseData = serData.map(function(d, i) {
                            var s = Object.create(d);
                            if (d.isDummy) {
                                dummyDataLength++;
                            }
                            if (stacked && animationType != HORIPATHTWEEN_ANIMATION) {
                                s.y0 = isPolar && animationType === STROKETWEEN_ANIMATION ? mathMin(0, scaleObj.domain()[0]) : 0;
                                s.y =
                                    !isPolar && hasNegValue != null && scaleObj(hasNegValue) != null
                                        ? mathMax(hasNegValue, scaleObj.domain()[0])
                                        : scaleObj.domain()[0];
                            } else {
                                if (isRangeChart && animationType == VERTPATHTWEEN_ANIMATION) {
                                    var yArr = new Array(2);
                                    var y0 = chartObj.dataset.getY(d, yaxiscolumnorder, 0);
                                    yArr[0] = y0;
                                    yArr[1] =
                                        s[idx] == null
                                            ? null
                                            : hasNegValue != null && scaleObj(hasNegValue) != null
                                            ? mathMax(hasNegValue, y0)
                                            : y0;
                                    s[idx] = yArr;
                                } else {
                                    s[idx] =
                                        s[idx] == null
                                            ? null
                                            : !isPolar && hasNegValue != null && scaleObj(hasNegValue) != null
                                            ? mathMax(hasNegValue, scaleObj.domain()[0])
                                            : scaleObj.domain()[baseIdx];
                                }
                            }
                            return s;
                        });
                    }
                    basedataShift = d3_values(baseData);
                }

                var ishideable = currentdata.disabled;
                var subseries = appendEle(
                    series,
                    'g',
                    function(d) {
                        return ishideable ? [] : [d];
                    },
                    '.',
                    'class',
                    `subseries_${i}`
                );
                var path = subseries.selectAll('path.area').data(function(d) {
                    return ishideable ? [] : [d];
                });

                var pathEnter = path
                    .enter()
                    .append('path')
                    .datum(baseData, function(d) {
                        return dataset.getX(d);
                    })
                    .attr('class', 'area')
                    .style('opacity', fillOpacity)
                    .style('fill', color)
                    .attr('d', area[i]);

                if (isAnimationEnable) {
                    if (animationType == POINTBYPOINT_ANIMATION || animationType == POINTTWEEN_ANIMATION) {
                        interTween(
                            baseData,
                            basedataShift,
                            serData,
                            categories,
                            animationType,
                            easingType,
                            transitionDuration,
                            rp,
                            function(d) {
                                return {
                                    d: area[i](d)
                                };
                            },
                            pathEnter
                        );
                    } else if (pathEnter.node() != null && animationType == STROKETWEEN_ANIMATION) {
                        //For stacked chart, check again
                        pathEnter
                            .transition()
                            .duration(0)
                            .delay(transitionDuration)
                            .attr('d', function(d) {
                                return area[i](d);
                            });
                    } else {
                        let isEmpty = !serData.length || (serData.length == 1 && !serData[0].length);
                        isEmpty = stacked ? serData.length == dummyDataLength : isEmpty;
                        if (isEmpty) {
                            pathEnter.attr('d', function(d) {
                                return area[i]([]);
                            });
                        } else {
                            pathEnter
                                .transition()
                                .duration(transitionDuration)
                                .ease(easingType)
                                .attrTween('d', pathTween);
                        }
                    }
                }

                function pathTween() {
                    var interpolate = d3_interpolate(area[i](baseData), area[i](serData));
                    return function(t) {
                        return interpolate(t);
                    };
                }
                var gradientDim = gradient_helpers.getDimension(chartObj, index);
                var fillColor = (chart.fillColorFunc = getFillColor(chartObj, index, paneid, gradientDim, null, null, null, rp));

                // var ShiftedData = d3_values(serData);
                // var ShiftedData_bef = [];
                if (shiftInfo != null) {
                    // ShiftedData = d3_values(serData);
                    // var co = shiftInfo.count;
                    // for (var k = 0; k <= co; k++) {
                    //     ShiftedData.splice(k, 0, shiftInfo.point)
                    // }
                    //
                    // for (var h = 0; h < serData.length; h++) {
                    //     ShiftedData_bef[h] = []
                    //     ShiftedData_bef[h][0] = serData[h][0]
                    //     ShiftedData_bef[h][1] = (h == 0) ? shiftInfo.point[1] : serData[h - 1][1]
                    // }
                    // ShiftedData_bef.push(["dummmy", serData[serData.length - 1][1]])

                    if (redraw) {
                        // var markerWidth = (chartObj.dataObject.totalnoofdatas <= maxLimit) && markerEnabled
                        var clipX = xscale(serData[0][0]) + (xscale.bandwidth != null ? xscale.bandwidth() / 2 : 0);
                        var clipWid =
                            chartObj.dataObject.getXYvalue(
                                axisrotated,
                                stacked,
                                xscale,
                                yscale,
                                yaxiscolumnorder,
                                serData[serData.length - 1],
                                'x'
                            ) - chartObj.dataObject.getXYvalue(axisrotated, stacked, xscale, yscale, yaxiscolumnorder, serData[0], 'x');

                        var properties = {
                                name: 'rect',
                                id: chartObj.id,
                                x: clipX,
                                y: 0,
                                width: clipWid,
                                height: chartObj.plotarea.height
                            },
                            args = {
                                chartObj: chartObj
                            };
                        var clipID = effects.get(effects.svg_clips, 'rect', properties, args);
                        series.attr('clip-path', `url(${defsLocation}#${clipID})`);
                    }
                }

                var dataUpdate = serData;
                var pathUpdate = pathEnter
                    .merge(path)
                    .datum(serData, function(d) {
                        return dataset.getX(d);
                    })
                    .styles({
                        opacity: fillOpacity,
                        fill: fillColor(legendOrder, null, null, gvp('fillColor') || color, NULL, NULL, NULL, NULL, i)
                    });

                if (shiftInfo != null) {
                    var transX =
                        shiftInfo != null
                            ? Math.abs(
                                  chartObj.dataObject.getXYvalue(
                                      axisrotated,
                                      stacked,
                                      chart.oldXscale,
                                      yscale,
                                      yaxiscolumnorder,
                                      serData[0],
                                      'x',
                                      null,
                                      yrangePos
                                  ) -
                                      chartObj.dataObject.getXYvalue(
                                          axisrotated,
                                          stacked,
                                          xscale,
                                          yscale,
                                          yaxiscolumnorder,
                                          serData[0],
                                          'x',
                                          null,
                                          yrangePos
                                      )
                              )
                            : null;
                    path.datum(dataUpdate)
                        .attr('d', area[i])
                        .attr('transform', null)
                        .transition()
                        .duration(transitionDuration)
                        .ease(easingType)
                        .attr('transform', function() {
                            return transX == null ? null : `translate(-${transX},0)`;
                        });
                } else {
                    var race_syncTrans = chartObj.racing && chartObj.racing.syncTransition;
                    if (pathEnter.node() != null && animationType == STROKETWEEN_ANIMATION) {
                        if (race_syncTrans) {
                            pathUpdate
                                .datum(dataUpdate)
                                .style('fill-opacity', 0)
                                .transition(race_syncTrans)
                                .attr('d', area[i])
                                .style('fill-opacity', null);
                        } else {
                            pathUpdate
                                .datum(dataUpdate)
                                .style('fill-opacity', 0)
                                .transition()
                                .duration(0)
                                .delay(transitionDuration)
                                .attr('d', area[i])
                                .style('fill-opacity', null);
                        }
                    } else {
                        if (race_syncTrans) {
                            pathUpdate
                                .datum(dataUpdate)
                                .style('fill-opacity', null)
                                .transition(race_syncTrans)
                                .attr('d', area[i]);
                        } else {
                            pathUpdate
                                .datum(dataUpdate)
                                .style('fill-opacity', null)
                                .transition()
                                .delay(
                                    (animationType == POINTBYPOINT_ANIMATION || animationType == POINTTWEEN_ANIMATION) &&
                                        pathEnter.node() != null
                                        ? transitionDuration
                                        : 0
                                )
                                .duration(transitionDuration)
                                .ease(easingType)
                                .attr('d', area[i]);
                        }
                    }
                }
                setOpacityForTickFilterElements(pathUpdate, chartObj, cType);

                if (chartObj.dataObject.hierarchical) {
                    pathUpdate.attr('levelbycolumn', function(d, i) {
                        return d[0] && d[0].levelByColumn;
                    });
                    if (chartObj.dataObject.pivot) {
                        pathUpdate.attr('levelbyrow', function(d, i) {
                            return d[0] && d[0].levelByRow;
                        });
                    }
                }
                path.exit().remove();
            });

            chart.getSymbolSize = line && line.getSymbolSize;
            if (!(borderEnabled || markerEnabled)) {
                if (isRangeChart) {
                    labelQueue.push({
                        renderer: chart,
                        charttype: 'scatter',
                        chartObj: chartObj,
                        paneid: paneid,
                        args: [currentdata, legendOrder, 0, 1]
                    });
                    labelQueue.push({
                        renderer: chart,
                        charttype: 'scatter',
                        chartObj: chartObj,
                        paneid: paneid,
                        args: [currentdata, legendOrder, 0, 0]
                    }); //last argument in args for yrangePos
                } else {
                    labelQueue.push({
                        renderer: chart,
                        charttype: 'scatter',
                        chartObj: chartObj,
                        rendererIndex: rendererIndex,
                        paneid: paneid,
                        args: [currentdata, legendOrder, 0]
                    });
                }
            }
        });
        let chartInstance = commonRendererProp.params.chartObj;
        chartInstance.timerObj[chartInstance.timerObj.length] = chart.renderedTimer = setTimeout(function() {
            chart.chartRendered = true;
            redraw = true;
        }, transitionDuration);
        chart.oldXscale = xscale;
        return chart;
    }
    chart.getHighlighter = function(chartObj, index) {
        if (!chart.highlighter) {
            chart.highlighter = chartObj.dataObject.isPolarAxisCategory
                ? new WebHighlighter(chartObj, chart)
                : new AreaHighlighter(chartObj, chart);
        }
        return chart.highlighter;
    };

    chart.commonRendererProp = function(_) {
        if (!arguments.length) {
            return commonRendererProp;
        }
        commonRendererProp = _;
        return chart;
    };

    chart.shiftInfo = function(_) {
        if (!arguments.length) {
            return shiftInfo;
        }
        shiftInfo = _;
        return chart;
    };

    chart.destroy = function() {
        line && line.destroy();
        baseLine && baseLine.destroy();
    }

    return chart;
};

export var _areaChart = svg_plot_renderer.area;
