// @zohocharts/zc/src/area/canvasAreachart.js
//$Id$
//canvasAreachart.js
import { utils, defined, pInt, isUndefined, mergeJSON } from '../utils/zcutils';
import { ORDINAL_DATATYPE } from '../core/datatype-constants';
import { getFillColor, getStrokeColor } from '../utils/colorutils';
import { gradient_helpers } from '../effects/gradient-helpers';
import { d3_values } from '../utils/zcutils';
import { getAreaLayout } from './area-layout';
import { labelRenderer, canvas_plot_renderer } from '../core/main-zc-init';
import { effects } from '../effects/zceffects';
import { AreaHighlighter } from './area-highlight';
import { WebHighlighter } from './web-highlight';
import { DataProcessor } from '../core/data-processor';
import { getChartTypeByIndex } from '../core/charttype-helpers';
import { getImageByIndex } from '../helpers/image-pallete-helpers';
import { isFacet } from '../axis/hierarchy-axis-utils';

canvas_plot_renderer.area = function() {
    var commonRendererProp,
        shiftInfo,
        shiftDiff,
        line = canvas_plot_renderer.line?.(),
        baseLine = canvas_plot_renderer.line?.();

    function chart(context, seriesdata, pattern, filterIndex) {
        var rp = commonRendererProp.params,
            chartObj = rp.chartObj,
            xscale = rp.xscale,
            yscale = rp.yscale,
            stacked = rp.stacked,
            index = rp.datasetIndex,
            legendOrder = rp.legendOrderIndex,
            rendererIndex = rp.rendererIndex,
            axisrotated = rp.axisrotated,
            commonPlotOptions = rp.commonPlotOptions,
            color = rp.color,
            labelQueue = rp.labelQueue,
            yaxiscolumnorder = rp.yaxiscolumnorder,
            mainSeries = seriesdata;

        seriesdata = chartObj.dataObject.categoryHierarchy ? seriesdata[index] : chartObj.seriesdata[index];
        var plotoptions = seriesdata.plotoptions;
        var paneid = isFacet(chartObj) ? `${seriesdata.row}_${seriesdata.column}` : '0_0';
        var gvp = (commonRendererProp.params.gvp = utils.getValueOfPlotOptions(plotoptions, commonPlotOptions));
        var fillOpacity = gvp('fillOpacity'),
            strokeWidth = gvp('strokeWidth'),
            strokeColor = gvp('strokeColor'),
            markeroptions = (commonRendererProp.params.markerOptions =
                plotoptions && plotoptions.marker
                    ? mergeJSON(commonPlotOptions.marker, plotoptions.marker, true)
                    : commonPlotOptions.marker),
            hoverOptions = (commonRendererProp.params.hoverOptions =
                plotoptions && plotoptions.hoveroptions
                    ? mergeJSON(commonPlotOptions.hoveroptions, plotoptions.hoveroptions, true)
                    : commonPlotOptions.hoveroptions),
            markerEnabled = markeroptions.enabled,
            borderEnabled =
                (defined(strokeWidth) && strokeWidth != '' && pInt(strokeWidth) != 0) ||
                (defined(strokeColor) && strokeColor != '') ||
                markerEnabled,
            datatypeX = chartObj.dataset.getDataType('x'),
            datatypeY = chartObj.dataset.getDataType('y', yaxiscolumnorder),
            xCategories = datatypeX == ORDINAL_DATATYPE ? xscale.domain() : null,
            yCategories = datatypeY == ORDINAL_DATATYPE ? yscale.domain() : null,
            cType_Obj = getChartTypeByIndex(chartObj, paneid, index),
            cType = cType_Obj.name,
            isRangeChart = (rp.isRangeChart = cType == 'arearange'),
            gradientDim = gradient_helpers.getDimension(chartObj, index),
            fillStyle = (chart.fillColorFunc = getFillColor(chartObj, legendOrder, paneid, gradientDim, null, null, null, rp)),
            strokeColor = getStrokeColor(chartObj, legendOrder);

        const gradient = pattern && pattern.gradient;
        if (defined(pattern)) {
            pattern = pattern.pattern || pattern;
            fillStyle = pattern;

            if (defined(filterIndex)) {
                markerEnabled = true;
                fillOpacity = 0.00000001;
            }
        }
        chart.layout = [];

        chart.seriesdataAfterUpdate = [];
        seriesdata.data.forEach(function(d, i) {
            var serData;
            if (datatypeX == ORDINAL_DATATYPE || datatypeY == ORDINAL_DATATYPE) {
                serData = chartObj.dataObject.updateSeriesdatabyCategories(d3_values(d), xCategories, yCategories, yaxiscolumnorder);
            } else {
                serData = d;
            }
            if (chartObj.renderVisiblePointsOnly || shiftInfo != null) {
                //for pushdatapoint allow filter for highlight
                serData = DataProcessor.helpers.filterPointsInViewport(chartObj, serData, 'area', 'x', index);
            }

            chart.seriesdataAfterUpdate[i] = serData;
            shiftDiff =
                shiftInfo != null
                    ? shiftDiff == null || shiftDiff == 0
                        ? chartObj.dataObject.getXYvalue(
                              axisrotated,
                              stacked,
                              xscale,
                              yscale,
                              yaxiscolumnorder,
                              serData[serData.length - 1],
                              'x'
                          ) -
                          chartObj.dataObject.getXYvalue(
                              axisrotated,
                              stacked,
                              xscale,
                              yscale,
                              yaxiscolumnorder,
                              serData[serData.length - 2],
                              'x'
                          )
                        : shiftDiff
                    : 0;
            var area = getAreaLayout(rp, shiftDiff, shiftInfo, serData, i, paneid),
                isSelection = context.canvas.className == 'selection-canvas' ? true : false;
            chart.layout[i] = area;
            if (!seriesdata.disabled) {
                var drawAreaFunction = function(context, fillStyle) {
                    context.beginPath();
                    context.fillStyle = fillStyle;
                    area.context(context)(d);
                    context.fill();
                    context.restore();
                };

                if (chartObj.imagePallete) {
                    effects.canvas_patterns.imageFill(
                        {
                            context: context,
                            name: getImageByIndex(chartObj, legendOrder, paneid)
                        },
                        drawAreaFunction
                    );
                } 
                else if(isSelection){
                    var pointsClicked = chartObj.dataObject.totalnoofdatas != chartObj.cache.selection.selectedPoints.length;

                    var fillStyle =
                    defined(pattern) && isUndefined(filterIndex) && !isSelection
                        ? pattern
                        : chart.fillColorFunc(
                              legendOrder,
                              null,
                              null,
                              isSelection && pointsClicked ?  pattern : gvp('fillColor') || color,
                              fillOpacity,
                              gradientDim,
                              null,
                              false,
                              i
                          );
                    gradient ? drawAreaFunction(context, gradient) : null;
                    drawAreaFunction(context, fillStyle);
                }
                else {
                    var fillStyle =
                        defined(pattern) && isUndefined(filterIndex)
                            ? pattern
                            : chart.fillColorFunc(
                                  legendOrder,
                                  null,
                                  null,
                                  gvp('fillColor') || color,
                                  fillOpacity,
                                  gradientDim,
                                  null,
                                  false,
                                  i
                              );
                    gradient ? drawAreaFunction(context, gradient) : null;
                    drawAreaFunction(context, fillStyle);
                }
            }
        });
        if (!seriesdata.disabled) {
            if (borderEnabled || markerEnabled) {
                line.commonRendererProp(commonRendererProp).shiftInfo(shiftInfo);
                line.chartRendered = false;
                if (isRangeChart) {
                    line.yrangePos(1);
                    baseLine.commonRendererProp(commonRendererProp).shiftInfo(shiftInfo).yrangePos(0);
                    baseLine.chartRendered = false;
                    baseLine(context, mainSeries, pattern, filterIndex);
                }
                line(context, mainSeries, pattern, filterIndex);
            }
        }

        if (defined(labelRenderer) && !(borderEnabled || markerEnabled)) {
            if (isRangeChart) {
                labelQueue.push({
                    renderer: chart,
                    charttype: 'scatter',
                    chartObj: chartObj,
                    rendererIndex: rendererIndex,
                    paneid,
                    args: [seriesdata, index, 0, 1]
                });
                labelQueue.push({
                    renderer: chart,
                    charttype: 'scatter',
                    chartObj: chartObj,
                    rendererIndex: rendererIndex,
                    paneid,
                    args: [seriesdata, index, 0, 0]
                });
            } else {
                labelQueue.push({
                    renderer: chart,
                    charttype: 'scatter',
                    chartObj: chartObj,
                    rendererIndex: rendererIndex,
                    paneid,
                    args: [seriesdata, index, 0, null]
                });
            }
        }
        chart.chartRendered = true;
        return chart;
    }

    chart.getHighlighter = function(chartObj, index) {
        if (!chart.highlighter) {
            chart.highlighter = chartObj.dataObject.isPolarAxisCategory
                ? new WebHighlighter(chartObj, chart)
                : new AreaHighlighter(chartObj, chart);
        }
        return chart.highlighter;
    };

    chart.commonRendererProp = function(_) {
        if (!arguments.length) {
            return commonRendererProp;
        }
        commonRendererProp = _;
        return chart;
    };

    chart.shiftInfo = function(_) {
        if (!arguments.length) {
            return shiftInfo;
        }
        shiftInfo = _;
        return chart;
    };
    return chart;
};

export var _canvasAreaChart = canvas_plot_renderer.area;
