// @zohocharts/zc/src/area/polyPoints.js
import { DataProcessor } from '../core/data-processor';
import { charttype } from '../core/main-zc-init';
import { defined } from '../utils/zcutils';

export function getPolygonPoints(data, point) {
    var prevpt,
        curpt,
        nextpt,
        prevBasept,
        curBasept,
        nextBasept,
        points = [];
    var itemIndex = defined(point.counter) ? point.counter : point.itemIndex,
        first = itemIndex == 0,
        last = itemIndex == data.length - 1,
        empty = [0, 0],
        current = point._data(),
        previous = first ? empty : data[itemIndex - 1]._data(),
        next = last ? empty : data[itemIndex + 1]._data();

    /*
    For area chart with null data points, we will break the data and show as different polygons.
    Hence we will check case for previous and next null data points. Construct polygon based on that.
    For normal data, polygon structure will be like, mid(prev, cur) -> cur -> mid(cur, next) and form the base data similarly.
    If previous is null point, then polygon will be like, cur -> mid(cur, next) . hence we will assign prev point to cur point
    If next is null point, then polygon will be like, mid(prev, cur) -> cur,. hence we will assign next point to cur point


    For stacked area with Previous or Next point is DUMMY, we need not find mid value. We will connect with previous point itself (dummy).
     */
    var isNull_current = !current._defined,
        isNull_previous = first ? true : !previous._defined,
        isNull_next = last ? true : !next._defined;

    var isDummy_previous = previous.isDummy,
        isDummy_next = next.isDummy;

    curpt = isNull_current ? empty : getPoint(current);
    prevpt = first ? empty : isNull_previous ? curpt : getPoint(previous);
    nextpt = last ? empty : isNull_next ? curpt : getPoint(next);

    curBasept = isNull_current ? empty : getBasePoint(current);
    prevBasept = first ? empty : isNull_previous ? curBasept : isDummy_previous ? prevpt : getBasePoint(previous);
    nextBasept = last ? empty : isNull_next ? curBasept : isDummy_next ? nextpt : getBasePoint(next);

    var prevCur = first ? empty : isNull_previous || isDummy_previous ? prevpt : getMidPoint(prevpt, curpt);
    var curNext = last ? empty : isNull_next || isDummy_next ? nextpt : getMidPoint(nextpt, curpt);

    var prevbaseCur = first ? empty : isNull_previous || isDummy_previous ? prevBasept : getMidPoint(prevBasept, curBasept);
    var curbaseNext = last ? empty : isNull_next || isDummy_next ? nextBasept : getMidPoint(nextBasept, curBasept);

    var temp =
        isNull_previous && isNull_next
            ? points.push(curpt)
            : first || isNull_previous
            ? points.push(curpt, curNext, curbaseNext, curBasept)
            : last || isNull_next
            ? points.push(prevCur, curpt, curBasept, prevbaseCur)
            : points.push(prevCur, curpt, curNext, curbaseNext, curBasept, prevbaseCur);
    return {
        points: points,
        prevCur: prevCur,
        curNext: curNext,
        curpt: curpt
    };
}
function getPoint(point) {
    return [point._x, point._y];
}
function getBasePoint(point) {
    return [point._x0, point._y0];
}
function getMidPoint(point1, point2) {
    return [(point1[0] + point2[0]) / 2, (point1[1] + point2[1]) / 2];
}

export function getAbstractPolyPoints(points) {
    let poly = [];
    points.filter(d => {
        var data = d._data();
        if (d._defined) {
            poly.push([d._x, d._y]);
        }
    });
    return poly;
}
