// @zohocharts/zc/src/area/web-highlight.js
import { _super, defined, inherit } from '../utils/zcutils';
import { ScatterHighlighter } from '../scatter/scatterHighlight';
import { getAbstractPolyPoints, getPolygonPoints } from './polyPoints';
import { patternFillEffect } from '../events/visualEffects/highlight/pattern-fill';
import { HIGHLIGHT_CONTEXT } from '../events/visualEffects/highlight/highlight-context';
import { AreaHighlighter } from './area-highlight';
import { d3_max, d3_merge, d3_min } from '../core/d3-constants';
import { appendEle } from '../utils/zcsvgutils';
import { effects } from '../effects/zceffects';
import { isPointVisible } from '../events/events-utils';
import { NULL } from '../core/zc-constants';

export function WebHighlighter(chart, renderer) {
    this.chart = chart;
    this.renderer = renderer;
}

inherit(WebHighlighter, ScatterHighlighter);

var proto = WebHighlighter.prototype;

proto._defaultH = function() {
    return [
        {
            shape: 'marker',
            path: _super(this).path,
            data: _super(this).data,
            context: _super(this).renderingContext,
            translate: this.translateContext
        },
        {
            shape: 'poly',
            path: this.path,
            data: this.data,
            context: this.renderingContext,
            translate: this.translateContext,
            styles: point => {
                return patternFillEffect.call(this, point, 'inherit');
            }
        }
    ];
};

proto.data = function(point) {
    var rp = this.renderer.commonRendererProp().params;
    var chart = this.chart,
        renderer = this.renderer,
        layout = renderer.layout[point.subSeriesIndex],
        y0 = layout.innerRadius(),
        isDefined = layout._defined;

    point._polydata = this.poly(point);

    return function() {
        if (!chart.cache.data[point._key]) {
            chart.cache.data[point._key] = {};
        }
        var cache = chart.cache.data[point._key];
        cache._x = point._x =
            cache._x ||
            chart.dataObject.getXYvalue(rp.axisrotated, rp.stacked, rp.xscale, rp.yscale, rp.yaxiscolumnorder, point, 'x') + rp.bound.x;
        cache._y = point._y =
            cache._y ||
            chart.dataObject.getXYvalue(rp.axisrotated, rp.stacked, rp.xscale, rp.yscale, rp.yaxiscolumnorder, point, 'y') + rp.bound.y;
        cache._y0 = point._y0 = cache._y0 || y0(point, point.itemIndex);
        cache._uid = point._uid = chart.dataset.getX(point);
        cache._defined = point._defined = defined(cache._defined) ? cache._defined : isDefined(point);
        cache._visible = point._visible = isPointVisible(chart, point);
        cache._point = point;
        return point;
    };
};

proto.poly = function(point) {
    var chart = this.chart,
        renderer = this.renderer;
    return function() {
        if (!chart.cache.data[point._key]) {
            chart.cache.data[point._key] = {};
        }
        var cache = chart.cache.data[point._key];
        if (!cache._poly) {
            let flatten = renderer.seriesdataAfterUpdate[point.subSeriesIndex],
                poly = getPolygonPoints(flatten, point),
                polypoints = getAbstractPolyPoints(flatten);
            poly.points = polypoints;
            cache._poly = point._poly = poly;
        }

        return point._poly;
    };
};

proto.renderingContext = function() {
    return HIGHLIGHT_CONTEXT.get('area').call(this, this.chart);
};

proto.translateContext = function() {
    return [this.chart.plotarea.left + this.chart.plotarea_approx.x, this.chart.plotarea.top + this.chart.plotarea_approx.y];
};

proto.options = function(point) {
    var rp = this.renderer.commonRendererProp().params,
        commonOptions = _super(this).commonOptions.call(this, point),
        symbol = commonOptions.symbol,
        size = commonOptions._size || rp.gvp('strokeWidth'),
        interpolation = rp.gvp('mode'),
        fillopacity = rp.hoverOptions.fillOpacity || 0.2;

    commonOptions.size = {
        outer: size * 0.7,
        inner: size * 0.35
    };
    commonOptions.fillopacity = fillopacity;
    commonOptions.interpolation = interpolation;

    return commonOptions;
};

proto.path = function(point, model) {
    var options = this.options(point),
        context = model.context.call(this),
        translate = model.translate.call(this),
        pathContext,
        rp = this.renderer.commonRendererProp().params,
        area = this.renderer.layout[point.subSeriesIndex];
    area.context(NULL);
    context.style('display', '').attr('transform', `translate(${translate[0] + rp.bound.x},${translate[1] + rp.bound.y})`);

    pathContext = appendEle(context, 'path', [1], '.', 'class', `highlight${point.seriesIndex}`);
    pathContext
        .datum(this.renderer.seriesdataAfterUpdate[point.subSeriesIndex])
        .attr('d', area)
        .attr('stroke', '')
        .styles({
            'fill-opacity': options.fillopacity,
            display: ''
        });

    return pathContext;
};
