// @zohocharts/zc/src/area/webClippath.js
import { effects } from '../effects/zceffects';
import { CIRCLE_ELEMENT } from '../core/element-constants';
import { mathMax } from '../core/zc-constants';
import { pointerEventsNonePolyfill } from '../events/eventActions';
import { getGridList } from '../core/data-processor/pivot-layout-helpers';
import { renderer_helpers, chartcategory } from '../core/main-zc-init';
import { isFacet } from '../axis/hierarchy-axis-utils';
import { arrayIncludes } from '../utils/zcutils';
import { isCategoryOf } from '../core/charttype-helpers';

export function addClipToTracker(chart) {
    let clipID;
    if (isFacet(chart)) {
        let desc = '',
            gridList = getGridList(chart.dataObject);

        gridList.forEach(grid => {
            let bound = renderer_helpers.getBound(chart, grid.data[0]),
                { x, y, width, height, polar } = bound,
                chartType = chart.dataObject.hierarchical.gridOptions[`${grid.row}_${grid.column}`].chartTypes[0].value;

            if (isCategoryOf(chart, chartType, 'polarAxis')) {
                let r = polar.radius;
                x += polar.x;
                y += polar.y;

                desc += `M${x - r} ${y} A${r} ${r} 0 0 1 ${x + r} ${y} A${r} ${r} 0 0 1 ${x - r} ${y} `;
            } else if (arrayIncludes(chartcategory.singleSourceEvent, chartType)) {
                desc += `M${x} ${y} L${x + width} ${y} L${x + width} ${y + height} L${x} ${y + height} Z `;
            }
        });

        var properties = { d: desc, name: 'pivotpolarClip', id: chart.id };

        clipID = effects.get(effects.svg_clips, 'polygon', properties, { chartObj: chart });
    } else if (isCategoryOf(chart, null, 'polarAxis')) {
        var plotarea_approx = chart.plotarea_approx,
            properties = {
                name: 'polarClip',
                id: chart.id,
                // #ZC2280
                cx: plotarea_approx.x,
                cy: plotarea_approx.y,
                r: mathMax(0, plotarea_approx.radius)
            },
            args = {
                chartObj: chart
            };

        clipID = effects.get(effects.svg_clips, CIRCLE_ELEMENT, properties, args);
    }

    chart.eventHandler.trackerElement.attr('clip-path', clipID ? `url(${chart.defsLocation}#${clipID})` : null);

    if (!window.PointerEvent && chart.plot.renderer.mode === 'canvas') {
        //ZC1533 pointer-events: none is not working in ie9, ie10
        pointerEventsNonePolyfill(chart.container.selectAll('canvas.zc-labelholder-canvas,canvas.highlight-canvas,canvas.zc-canvas'));
    }
}
