// @zohocharts/zc/src/area/webOptions.js
import { defaultTheme as options } from '../core/defaulttheme';
options.web = function() {
    return {
        fillOpacity: 0.5,
        strokeWidth: 2,
        mode: 'linearClosed', //linearClosed || cardinalClosed
        outerPadding: 10,
        stacked: {
            enabled: false,
            showAs: 'values' //values,percent
        },
        marker: {
            enabled: true,
            innerStrokeOpacity: 1,
            outerFillColor: 'white',
            outerStrokeWidth: 0,
            useSVGMarkers: null,
            sizeRatio: null
        },
        hoveroptions: {
            innerStrokeWidth: 2,
            innerStrokeOpacity: 1,
            outerFillColor: 'white'
        },
        datalabels: {
            showAs: 'y' //x,y,percent
        },
        animation: {
            type: 'strokeTween', //strokeTween , pathTween , pointbypointTween , fade
            marquee: {
                enabled: false,
                duration: 300
            },
            blink: {
                enabled: false,
                type: 'twinkle', //twinkle | fade
                duration: 800
            }
        },
        gradients: {
            // radial || linear || none
            options: {
                linear: {
                    x2: 15,
                    spreadMethod: 'repeat', // pad || repeat || reflect
                    gradientUnits: 'userSpaceOnUse' //objectBoundingBox || userSpaceOnUse
                },
                radial: {
                    radius: 90,
                    x: 0,
                    y: 0
                }
            }
        }
    };
};

export var _webOptions = options.web;
