// @zohocharts/zc/src/pyramid/pyramidchart.js
//$Id$
//pyramidchart.js
import { Registry } from '../core/zc-registry';
import { legend_helpers as L_H, svg_plot_renderer } from '../core/main-zc-init';
import { mathMin, mathMax, mathSQRT, mathPow, mathSin, mathPI, mathAbs } from '../core/zc-constants';
import { isDatalabelEnabled } from '../label/datalabelsUtils';
import { utils, validateNullVal, pInt, defined } from '../utils/zcutils';
import { getAnimationConf } from '../utils/animationutils';
import { d3_select, d3_scaleLinear, d3_ease_linear, d3_interpolate, d3_event, d3_line } from '../core/d3-constants';
import { appendEle } from '../utils/zcsvgutils';
import { getFillColor, getStrokeColor, hasColorColumnInNonAxis } from '../utils/colorutils';
import { PATH_ELEMENT } from '../core/element-constants';
import { setPointerEvents } from '../events/events-utils';
import { generateDataForAnnotation } from '../annotation/annotation-utils';
import { attachevents } from '../events/eventActions';
import { VERTICAL_ANIMATION } from '../core/animation-constants';
import { labelRenderer } from '../core/main-zc-init';
import { FUNNEL_CHARTNAME, PYRAMID_CHARTNAME } from '../core/chartname-constants';
import { getEventConf, getEventObject } from '../events/events-utils';
// import { legend_helpers as L_H } from '../legend/legend-helper';
import { updateSelectionOnDatapoint } from '../events/event-update-selection';
import { isFacet } from '../axis/hierarchy-axis-utils';
import { setOpacityForTickFilterElements, setPivotAttributes } from '../plot/svg/set-pivot-attr';
import { zcMorph } from '../morphing/zc-morping';
import { getRawColor } from '../utils/colorutils';
import { getOffsetValue } from '../plot/morph/morphTools';

let annotation = Registry.getComponent('annotation');
/**
 *
 * @returns {chart} pyramid renderer.
 */
svg_plot_renderer.pyramid = function() {
    var MAXWIDTH_DATALABELS = 0.3;
    var cfg = {};
    var animationType;
    var commonRendererProp,
        dataset,
        yaxiscolumnorder,
        isredraw = false;
    var chartWidth, chartHeight, commonPlotOptions, systemconf, notesArray, isPatternPalleteEnabled, defsLocation;
    var height1,
        total,
        b,
        h,
        side1,
        side2,
        side3,
        side4,
        hScale,
        color,
        heightArray,
        chartObj,
        pyramidGroup,
        gradientDim,
        pyramidsections,
        dataLength,
        data_backup,
        datalabelPadding,
        customPlotarea,
        fillColor = [];
    let shiftInfo;

    function chart(selection) {
        selection.each(function(data) {
            //renderer config values
            var rp = commonRendererProp.params;
            (chartObj = rp.chartObj),
                (commonPlotOptions = rp.commonPlotOptions),
                (color = rp.color),
                (notesArray = rp.notesArray),
                (yaxiscolumnorder = rp.yaxiscolumnorder),
                (dataset = rp.dataset);
            systemconf = chartObj.systemConf;
            var labelQueue = rp.labelQueue,
                { bound, rendererIndex } = rp,
                seriesData = isFacet(chartObj) && hasColorColumnInNonAxis(chartObj) ? data : data[0],
                paneid = isFacet(chartObj) ? `${seriesData.row}_${seriesData.column}` : null;

            defsLocation = chartObj.defsLocation;

            let plotarea = { ...bound, ...customPlotarea };
            chartWidth = plotarea.width;
            chartHeight = plotarea.height;
            var minD = mathMin(chartWidth, chartHeight);

            datalabelPadding = commonPlotOptions.datalabels.padding;
            var showlabel = chartObj.datalabels.moduleExists && isDatalabelEnabled(commonPlotOptions.datalabels, chartObj);
            if (showlabel) {
                /*
                 *If chart width is lesser than height, datalabels are hidden as default padding is 10 and there is no space to show datalabels.
                 * So 30% of the width is allocated for datalabels and the default padding will be changed to 5
                 */
                var minDatalabelsWidth = chartWidth * MAXWIDTH_DATALABELS,
                    availableDatalabelsWidth = chartWidth - minD;
                if (availableDatalabelsWidth < minDatalabelsWidth) {
                    datalabelPadding = cfg.datalabelPadding = 5;
                    minD -= mathMax(minDatalabelsWidth - availableDatalabelsWidth, 0);
                }
            }
            cfg.width = minD;
            cfg.height = (minD * 4) / 5;

            var plotoptions = data[0].plotoptions,
                gvp = utils.getValueOfPlotOptions(plotoptions, commonPlotOptions);
            isPatternPalleteEnabled = systemconf.legend.imagePallete || L_H.hasPatternPallete(chartObj);

            var strokeWidth = gvp('strokeWidth'),
                strokeOpacity = gvp('strokeOpacity'),
                threeDimension = gvp('threeDimension'),
                fillOpacity = gvp('fillOpacity');

            //Animation variables
            var animationOptions = getAnimationConf(commonPlotOptions, plotoptions, chartObj, PYRAMID_CHARTNAME);
            var isAnimationEnable = animationOptions.enabled,
                easingType = animationOptions.easingType,
                transitionDuration = isredraw ? (isAnimationEnable ? 200 : 0) : animationOptions.duration;
            animationType = animationOptions.type;

            var eventDisabled = validateNullVal(getEventConf(plotoptions, commonPlotOptions, chartObj, 'enabled'), true) == false;
            var cursor = eventDisabled ? null : getEventConf(plotoptions, commonPlotOptions, chartObj, 'cursor') || 'pointer';

            var transX = chartWidth / 2 - cfg.width / 2,
                transY = chartHeight / 2 - cfg.height / 2,
                translation = hasColorColumnInNonAxis(chartObj)
                    ? [transX + plotarea.translate[0], transY + plotarea.translate[1]]
                    : [transX, transY];
            (chartObj.plotarea.centerX = transX), (chartObj.plotarea.centerY = transY);
            var container = d3_select(this);
            pyramidGroup = appendEle(container, 'g', [data], '.', 'class', 'pyramidgroup');
            pyramidGroup.attr('transform', `translate(${transX},${transY})`).style('cursor', cursor);
            data_backup = data = hasColorColumnInNonAxis(chartObj) ? data : data[0].data[0];
            dataLength = data.length;
            heightArray = [];
            height1 = cfg.height;
            (b = cfg.width), (h = threeDimension ? cfg.height / 6 : 0);
            var theta = 25;
            var x = mathSQRT(mathPow(h / mathSin((theta * mathPI) / 180), 2) - mathPow(h, 2));
            b = b - x;
            side1 = d3_scaleLinear()
                .range([(b + x) / 2, 0])
                .domain([0, height1]);
            side2 = d3_scaleLinear()
                .range([(b + x) / 2, b])
                .domain([0, height1]);
            side3 = d3_scaleLinear()
                .range([(b + x) / 2, b + x])
                .domain([0, height1]);
            side4 = d3_scaleLinear()
                .range([(b + x) / 2, x])
                .domain([0, height1]);
            hScale = d3_scaleLinear()
                .range([0, h])
                .domain([0, height1 - h / 2]);
            gradientDim = [cfg.width, cfg.height];
            chart.fillColorFunc = getFillColor(chartObj, 0, paneid, gradientDim);
            var strokeColor = getStrokeColor(chartObj, 0, paneid);

            if (isFacet(chartObj)) {
                data.forEach((d, i) => {
                    let itemIndex = dataLength - 1 - i;
                    fillColor[itemIndex] = getFillColor(chartObj, itemIndex, paneid);
                });
            } else {
                fillColor[0] = chart.fillColorFunc;
            }

            var noofNonZeroData = 0;
            chart.total = total = data.reduce(function(prev, cur) {
                return prev + (isDisabled(cur) ? 0 : (noofNonZeroData++, mathAbs(dataset.getY(cur, yaxiscolumnorder) || 0)));
            }, 0);

            var previous = 0;
            data.forEach(function(d, i) {
                const index = dataLength - 1 - i;
                d._acutalData = () => data[index];
                heightArray.push({
                    height:
                        ((previous + (isDisabled(d) ? 0 : mathAbs(dataset.getY(d, yaxiscolumnorder)))) / total) * (height1 - h / 2) || 0,
                    previousHeight: (previous / total) * (height1 - h / 2) || 0
                });
                previous += isDisabled(d) ? 0 : mathAbs(dataset.getY(d, yaxiscolumnorder));
            });
            pyramidsections = appendEle(pyramidGroup, 'g', data, '.', 'class', 'pyramidsection');
            pyramidsections
                .attr('id', function(d, i) {
                    return `section_${dataLength - 1 - i}`;
                })
                .styles({
                    stroke: function(d, i) {
                        const index = dataLength - 1 - i;
                        if (isFacet(chartObj)) {
                            return getStrokeColor(chartObj, index, paneid)(
                                index,
                                data[index],
                                commonPlotOptions.strokeColor
                            );
                        }

                        return strokeColor(index, data[index], commonPlotOptions.strokeColor);
                    },
                    'stroke-opacity': strokeOpacity,
                    'stroke-width': strokeWidth,
                    'fill-opacity': fillOpacity,
                    fill: function(d, i) {
                        const index = dataLength - 1 - i;
                        if (isFacet(chartObj)) {
                            return fillColor[index](index, data[index]);
                        }

                        return fillColor[0](index, data[index]);
                    }
                })
                .each(function(d, i) {
                    var currentGroup = d3_select(this);
                    var pyramid = appendEle(currentGroup, PATH_ELEMENT, [d, d, d, d]);
                    pyramid.attr('class', `section_${dataLength - 1 - i}`); //For touch events, class will help to get index
                });

            setPivotAttributes(pyramidsections, chartObj);
            setOpacityForTickFilterElements(pyramidsections, chartObj, 'pyramid');

            setPointerEvents(chartObj, pyramidsections, eventDisabled, paneid);
            if (!chartObj.cache.search) {
                chartObj.cache.search = {};
            }
            if (!eventDisabled) {
                var mousemove = function(e, d) {
                        const ele = this.parentNode;
                        var args = getEventArgs.call(this, e, d);
                        const i = args.itemIndex;
                        chartObj.eventHandler.handlers.mousemove.call(chartObj.eventHandler, e, args);
                        updateSelectionOnDatapoint(data[i], args);
                        chartObj.cache.search.lastPane = paneid;
                    },
                    mouseout = function(e, d) {
                        const ele = this.parentNode;
                        var args = getEventArgs.call(this, e, d);
                        const i = args.itemIndex;
                        chartObj.eventHandler.handlers.mouseout.call(chartObj.eventHandler, e, args);
                        updateSelectionOnDatapoint(data[i], args);
                        chartObj.cache.search.lastPane = null;
                    },
                    click = function(e, d) {
                        const ele = this.parentNode;
                        var args = getEventArgs.call(this, e, d);
                        const i = args.itemIndex;
                        chartObj.eventHandler.handlers.click.call(chartObj.eventHandler, e, args);
                        updateSelectionOnDatapoint(data[i], args);
                        chartObj.cache.search.lastPane = paneid;
                    },
                    dblclick = function(e, d) {
                        var ele = this.parentNode;
                        var args = getEventArgs.call(this, e, d);
                        const i = args.itemIndex;
                        var dt = !annotation._empty ? generateDataForAnnotation(data[i], null) : data[i];
                        var height = heightArray[i].previousHeight;
                        var coordinates = [
                            side1(height + hScale(height) / 2) + (chartWidth / 2 - cfg.width / 2),
                            hScale(height) / 2 + height + (chartHeight / 2 - cfg.height / 2)
                        ];
                        var args = getEventArgs.call(this, e, d);
                        args.coordinate = coordinates;
                        args.basedOn = 'plot';
                        args.data = dt;
                        chartObj.eventHandler.handlers.doubleclick.call(chartObj.eventHandler, e, args);
                        updateSelectionOnDatapoint(data[i], args);
                        chartObj.cache.search.lastPane = paneid;
                    },
                    eventAttacherOptions = {
                        target: 'plot',
                        touchleave: true,
                        chartObj: chartObj
                    };
                var mouseHandlers = {
                        mousemove: mousemove,
                        mouseout: mouseout,
                        click: click,
                        dblclick: dblclick
                    },
                    touchHandlers = {
                        singletouchmove: mousemove,
                        touchend: mouseout,
                        tap: click,
                        dbltap: dblclick
                    };
                attachevents(pyramidsections.selectAll(PATH_ELEMENT), mouseHandlers, touchHandlers, eventAttacherOptions);
            }
            var prev = 0;
            pyramidsections.each(function(d, i) {
                var pyramid = pyramidGroup.selectAll(`#section_${i}`).selectAll(PATH_ELEMENT);
                if (isredraw) {
                    pyramid
                        .transition()
                        .duration(transitionDuration)
                        .ease(d3_ease_linear)
                        .attr('d', function(d, j) {
                            return getPath(i, j);
                        });
                } else {
                    var yVal = dataset.getY(d, yaxiscolumnorder);
                    var dur =
                        animationType == VERTICAL_ANIMATION
                            ? ((isDisabled(d) || !yVal ? 0 : mathAbs(yVal)) / total) * transitionDuration
                            : transitionDuration;
                    prev += isDisabled(d) || !yVal ? 0 : mathAbs(yVal);
                    var del = animationType == VERTICAL_ANIMATION ? ((total - prev) / total) * transitionDuration : 0;
                    if (chartObj.systemConf.chart.plot.morph.enabled && chartObj.oldSeriesInfo != undefined) {
                        var morph_Obj = new zcMorph({ sampling: chartObj.systemConf.chart.plot.morph.samplingDistance });
                        var oldseriesData = chartObj.oldSeriesInfo.get(0);
                        var oldseriesIndex = oldseriesData.seriesIndex;
                        var oldseriesColor = oldseriesData.color;
                        var offsetValue = getOffsetValue(chartObj, oldseriesIndex);
                        var individualElementTransform = oldseriesData.transformationInfo;
                        transitionDuration = chartObj.systemConf.chart.plot.morph.duration;

                        var height = heightArray[i].height;
                        var previous = heightArray[i].previousHeight;

                        var finalPath = `M${side1(previous + hScale(previous) / 2)} ${hScale(previous) / 2 + previous} L${side2(
                            previous + hScale(previous) / 2
                        )} ${hScale(previous) / 2 + previous} L${side3(previous + hScale(previous) / 2)} ${-hScale(previous) / 2 +
                            previous} L${side3(height + hScale(height) / 2)} ${-hScale(height) / 2 + height} L${side2(
                            height + hScale(height) / 2
                        )} ${hScale(height) / 2 + height} L${side1(height + hScale(height) / 2)} ${hScale(height) / 2 + height}`;

                        var line = d3_line();

                        var fromPath = chartObj.oldSeriesInfo.get(oldseriesIndex).pathData[i];
                        var fromPathAttribute = fromPath.getAttribute('d');
                        fromPath.setAttribute('d', finalPath);
                        var toPath = fromPath.cloneNode();
                        fromPath.setAttribute('d', fromPathAttribute);
                        var interpolated = morph_Obj.interpolate(
                            {
                                path: fromPath,
                                xOffset: offsetValue.x + individualElementTransform[i].x,
                                yOffset: offsetValue.y + individualElementTransform[i].y
                            },
                            { path: toPath, xOffset: 0, yOffset: 0 }
                        );

                        pyramid.each(function(d, j) {
                            if (j === 0) {
                                var pyramidSection = pyramidGroup.selectAll(`#section_${i}`);
                                pyramidSection
                                    .style('fill', function() {
                                        return oldseriesColor[i];
                                    })
                                    .transition()
                                    .duration(transitionDuration)
                                    .ease(easingType)
                                    .style('fill', function() {
                                        return getRawColor(chartObj, i, d.data, oldseriesIndex, null);
                                    });
                                d3_select(this)
                                    .transition()
                                    .duration(transitionDuration)
                                    .ease(easingType)
                                    .attrTween('d', function(d, i) {
                                        return function(t) {
                                            return line(
                                                interpolated.map(function(p) {
                                                    return p(t);
                                                })
                                            );
                                        };
                                    })
                                    .on('end', function(d, j) {
                                        if (chartObj.oldSeriesInfo) {
                                            delete chartObj.oldSeriesInfo;
                                        }
                                        if (j === 0) {
                                            pyramid.each(function(d, j) {
                                                d3_select(this).attr('d', getPath(i, j));
                                            });
                                            pyramidSection.style('fill', function() {
                                                if (isFacet(chartObj)) {
                                                    return fillColor[i](i, data[i]);
                                                }
                                                return fillColor[0](i, data[i]);
                                            });
                                        }
                                    });
                            }
                        });
                    } else {
                        pyramid
                            .transition()
                            .duration(dur)
                            .delay(del)
                            .ease(easingType)
                            .attrTween('d', function(dt, j) {
                                return tween(d, i, j);
                            });
                    }
                }
            });
            hideDisabledElements();

            chartObj.timerObj[chartObj.timerObj.length] = chart.renderedTimer = setTimeout(function() {
                if (labelRenderer) {
                    var args = [
                        chartObj.seriesdata,
                        0,
                        data,
                        noofNonZeroData,
                        0,
                        total,
                        cfg,
                        calculatePointsForLabelPlacement(data),
                        chart.subSeriesIndex || 0,
                        translation
                    ];
                    labelQueue.push({
                        renderer: chart,
                        charttype: FUNNEL_CHARTNAME,
                        chartObj: chartObj,
                        args: args,
                        rendererIndex,
                        paneid
                    });
                    labelQueue.push({
                        renderer: chart,
                        charttype: 'funnelInnerLabels',
                        chartObj: chartObj,
                        args: args,
                        rendererIndex,
                        paneid
                    });
                }

                //Annotation preprocessing
                if (!annotation._empty) {
                    notesArray = annotation.preprocessAnnotations(
                        chartObj.seriesdata,
                        0,
                        function(dt, k, j) {
                            var height = heightArray[j].previousHeight;
                            return [
                                side1(height + hScale(height) / 2) + (chartWidth / 2 - cfg.width / 2),
                                hScale(height) / 2 + height + (chartHeight / 2 - cfg.height / 2)
                            ];
                        },
                        notesArray,
                        chartObj,
                        rp
                    );
                }

                isredraw = true;
                chart.chartRendered = true;

                if (hasColorColumnInNonAxis(chartObj)) {
                    chartObj.renderer[rendererIndex].endFunctions();
                }
            }, transitionDuration + 10);

            function getEventArgs(e, d) {
                const ele = this.parentNode,
                    i = pInt(
                            d3_select(ele)
                                .attr('id')
                                .replace(/section_/gi, '')
                        );
                var ev = getEventObject(e);
                return {
                    renderer: chart,
                    seriesIndex: hasColorColumnInNonAxis(chartObj) ? i : 0,
                    subSeriesIndex: 0,
                    itemIndex: i,
                    paneid: paneid,
                    rendererIndex: rendererIndex,
                    color: color[i % color.length],
                    coordinates: [ev.pageX, ev.pageY],
                    element: ele,
                    data: data[i],
                    selected: data[i].selected,
                    _key: `ZC_${chartObj.id}_0_0_${i}`
                };
            }

            function hideDisabledElements() {
                if (!isFacet(chartObj)) {
                    return;
                }

                let dur = !isredraw ? 0 : transitionDuration;
                pyramidsections.each(function(d, i) {
                    var hide = d.dummy || !dataset.getY(d, yaxiscolumnorder);
                    var t = hide ? dur : 0;
                    if (t) {
                        setTimeout(function() {
                            hideSection(i, hide);
                        }, t);
                    } else {
                        hideSection(i, hide);
                    }
                });
            }

            function hideSection(i, hide) {
                var display = hide ? 'none' : null;

                pyramidGroup.selectAll(`#section_${i}`).style('display', display);
            }
        });
    }

    chart.commonRendererProp = function(_) {
        if (!arguments.length) {
            return commonRendererProp;
        }
        commonRendererProp = _;
        return chart;
    };

    chart.plotarea = function(_) {
        //Bubble Pie chart will give you the plotarea
        if (!arguments.length) {
            return customPlotarea;
        }
        customPlotarea = _;
        return chart;
    };

    chart.defaultMouseOverFunc = function(d, seriesPos, itemPos) {
        //Reset previous pane mouseover
        // if(chartObj.dataObject.pivot){
        //     var allPyramidSection = chartObj.wrap.selectAll('path.pyramidsection');
        //     if(isPatternPalleteEnabled){
        //         allPyramidSection.style('opacity', null);
        //     }else{
        //         allPyramidSection.selectAll('path').style('fill', (d,i)=>{
        //             return fillColor[0](dataLength - 1 - i, data_backup[dataLength - 1 - i], null);
        //         });
        //     }
        //
        // }

        if (isPatternPalleteEnabled) {
            pyramidsections.style('opacity', function(d, i) {
                return dataLength - 1 - i == itemPos ? null : 0.2;
            });
        } else {
            pyramidsections.style('fill', function(d, i) {
                return (isFacet(chartObj)
                    ? fillColor[dataLength - 1 - i]
                    : fillColor[0])(dataLength - 1 - i, data_backup[dataLength - 1 - i], dataLength - 1 - i == itemPos ? 0.5 : null);
            });
        }
    };
    chart.defaultMouseOutFunc = function(d, i) {
        if (isPatternPalleteEnabled) {
            pyramidsections.style('opacity', null);
        } else {
            pyramidsections.style('fill', function(d, i) {
                return (isFacet(chartObj)
                    ? fillColor[dataLength - 1 - i]
                    : fillColor[0])(dataLength - 1 - i, data_backup[dataLength - 1 - i]);
            });
        }
    };

    function constructSide1(height, previous) {
        return `M${side1(height + hScale(height) / 2)} ${hScale(height) / 2 + height} L${side2(height + hScale(height) / 2)} ${hScale(
            height
        ) /
            2 +
            height} L${side2(previous + hScale(previous) / 2)} ${hScale(previous) / 2 + previous} L${side1(
            previous + hScale(previous) / 2
        )} ${hScale(previous) / 2 + previous} L${side1(height + hScale(height) / 2)} ${hScale(height) / 2 + height}`;
    }

    function constructSide2(height, previous) {
        return `M${side2(height + hScale(height) / 2)} ${hScale(height) / 2 + height} L${side3(height + hScale(height) / 2)} ${-hScale(
            height
        ) /
            2 +
            height} L${side3(previous + hScale(previous) / 2)} ${-hScale(previous) / 2 + previous} L${side2(
            previous + hScale(previous) / 2
        )} ${hScale(previous) / 2 + previous} L${side2(height + hScale(height) / 2)} ${hScale(height) / 2 + height}`;
    }

    function constructSide3(height, previous) {
        return `M${side3(height + hScale(height) / 2)} ${-hScale(height) / 2 + height} L${side4(height + hScale(height) / 2)} ${-hScale(
            height
        ) /
            2 +
            height} L${side4(previous + hScale(previous) / 2)} ${-hScale(previous) / 2 + previous} L${side3(
            previous + hScale(previous) / 2
        )} ${-hScale(previous) / 2 + previous} L${side3(height + hScale(height) / 2)} ${-hScale(height) / 2 + height}`;
    }

    function constructSide4(height, previous) {
        return `M${side4(height + hScale(height) / 2)} ${-hScale(height) / 2 + height} L${side1(height + hScale(height) / 2)} ${hScale(
            height
        ) /
            2 +
            height} L${side1(previous + hScale(previous) / 2)} ${hScale(previous) / 2 + previous} L${side4(
            previous + hScale(previous) / 2
        )} ${-hScale(previous) / 2 + previous} L${side4(height + hScale(height) / 2)} ${-hScale(height) / 2 + height}`;
    }

    function getPath(i, j, base) {
        var height = heightArray[i].height;
        var baseHgt = base == 'down' ? height : heightArray[i].previousHeight;
        switch (j) {
            case 0:
                return constructSide3(height, baseHgt);
            case 1:
                return constructSide4(height, baseHgt);
            case 2:
                return constructSide1(height, baseHgt);
            case 3:
                return constructSide2(height, baseHgt);
        }
    }

    function tween(d, i, j) {
        var interpolate = d3_interpolate(getPath(animationType == VERTICAL_ANIMATION ? i : dataLength - 1, j, 'down'), getPath(i, j));
        return function(t) {
            return interpolate(t);
        };
    }

    function calculatePointsForLabelPlacement(data) {
        var trape = [];
        data.forEach(function(d, i) {
            var idx = data.length - 1 - i;
            var height = heightArray[idx].height,
                previous = heightArray[idx].previousHeight;
            var midHeight = (height + previous) / 2;
            var bottom_right = {
                    x: side3(height + hScale(height) / 2),
                    y: -hScale(height) / 2 + height
                },
                top_right = {
                    x: side3(previous + hScale(previous) / 2),
                    y: -hScale(previous) / 2 + previous
                },
                bottom_left = {
                    x: side1(height + hScale(height) / 2),
                    y: hScale(height) / 2 + height
                },
                top_left = {
                    x: side1(previous + hScale(previous) / 2),
                    y: hScale(previous) / 2 + previous
                };
            var rightX = (bottom_right.x + top_right.x) / 2,
                rightY = (bottom_right.y + top_right.y) / 2,
                leftX = (bottom_left.x + top_left.x) / 2,
                leftY = (bottom_left.y + top_left.y) / 2,
                midX = (side1(midHeight + hScale(midHeight) / 2) + side2(midHeight + hScale(midHeight) / 2)) / 2;

            trape.push({
                rightx: rightX + datalabelPadding,
                leftx: leftX - datalabelPadding,
                y: rightY,
                y0: height + hScale(height) / 2,
                midX: midX,
                leftY: leftY
            });
        });
        return trape;
    }

    function isDisabled(d) {
        return d.disabled || d.dummy;
    }

    chart.shiftInfo = function(_) {
        return arguments.length ? (shiftInfo = _, chart): shiftInfo;
    }

    chart.destroy = function destroy() {
        pyramidGroup = pyramidsections = null;
    }

    return chart;
};

export var _pyramidChart = svg_plot_renderer.pyramid;
