// @zohocharts/zc/src/pyramid/pyramidOptions.js
import { defaultTheme as options } from '../core/defaulttheme';
options.pyramid = function() {
    return {
        threeDimension: true,
        strokeColor: 'black',
        strokeWidth: null,
        strokeOpacity: 0.5,
        fillOpacity: 0.8,
        gradients: {
            // linear || none
            options: {
                linear: {
                    x2: 50,
                    spreadMethod: 'reflect', // pad || repeat || reflect
                    colorGamma: [0.8, 0]
                }
            }
        },
        datalabels: {
            showAs: 'y', //x,y,percent
            type: 'singleside', // singleside || doubleside || strokeddoubleside
            padding: 10,
            relativeX: true,
            line: {
                strokeColor: 'grey',
                strokeWidth: 0.3
            },
            innerLabel: {
                show: false,
                showAs: 'percent' //x,y,percent
            }
        },
        animation: {
            type: 'vertical' //vertical || verticalAll
        }
    };
};

export var _pyramidOptions = options.pyramid;
