// @zohocharts/zc/src/title/title-parser.js
//
import { BOTTOM_ALIGN, TEXT_ANCHOR_MAP } from '../core/align-constants';
import { Mark } from '../mark/mark-utils';
import { decodeHTML, bbox, pick, parseShortHandValue } from '../utils/zcutils';
import { trimText } from '../utils/zcsvgutils';

let DY_MAP = {
    top: '0.91em', // TODO: check '1em'
    center: '0.3em',
    bottom: '-0.3em'
};

function TitleParserSvg(spec) {
    let { config, data, bound } = spec,
        { text, hAlign = 'left', vAlign = 'top' } = config,
        margin = parseShortHandValue(pick(config.margin, 3)), // Have to check with spec.margin
        _margin = parseShortHandValue(pick(spec.margin, config.margin, 3)),
        _edge = edge(bound, margin),
        //
        preparedConfig = { text: decodeHTML(text), x: margin.left, y: margin.top, dy: DY_MAP[vAlign], anchor: TEXT_ANCHOR_MAP[hAlign] };

    let titleMark = Mark('text', null, data);
    titleMark.onEnd = _edge && _margin && (ele => trimTitle(ele, _edge, _margin));
    titleMark.config = { ...config, ...preparedConfig };

    return titleMark;
}

function trimTitle(ele, edge, margin) {
    let node = ele.node(),
        bb = bbox(node),
        pattern = '..',
        textBound = {
            left: 0,
            top: 0,
            width: bb.width,
            height: bb.height
        };

    trimText(node, node.textContent, textBound, edge, BOTTOM_ALIGN, pattern, margin);

    if (node.textContent === pattern) {
        ele.text('');
    }
}

function edge(bound, margin) {
    let { x = 0, y = 0 } = bound;

    return {
        left: x + margin.left,
        top: y + margin.top,
        right: x + bound.maxWidth + margin.left - margin.right,
        bottom: y + bound.maxHeight + margin.top - margin.bottom
    };
}

export { TitleParserSvg };
