// @zohocharts/zc/src/legend/events/legend-event-helpers.js
//
import { Registry } from '../../core/zc-registry';
import { LIST_LEGEND_TYPE, NESTED_LEGEND_TYPE, BAND_LEGEND_TYPE } from '../legend-type-constants';

export function isSupported(name) {
    return function(chart, id = 'z-0') {
        return getLegendView(chart.legends, id) === name;
    };
}

export function indexOfRange(d, ranges) {
    let index;

    ranges.some((_, i) => {
        let reversedIndex = ranges.length - 1 - i,
            range = ranges[reversedIndex];

        if (d >= range[0] && d <= range[1]) {
            index = reversedIndex;
            return true;
        }
    });

    return index;
}
Registry.setComponent('legend.event.helpers.indexOfRange', indexOfRange);

export function getLegendView(legends, id) {
    let sizeLegend = legends.get(id);

    return sizeLegend && sizeLegend.meta.view;
}

export function getHoverOptions(config, encode = 'z', byView = true) {
    let view = getViewByConfig(config, encode),
        { hoverOptions, itemHoverStyle, _itemHoverStyle } = config, // _itemHoverStyle: color band FEATURESEVENTS_HEATMAP_9_1
        _options = (encode === 'clr' ? _itemHoverStyle || itemHoverStyle : hoverOptions) || {};

    return (byView ? _options[view] : null) || _options;
}

export function getDisabledOptions(config, encode = 'z') {
    let view = getViewByConfig(config, encode),
        { filter, itemDisabledStyle } = config,
        _options = (encode === 'clr' ? itemDisabledStyle : filter) || {};

    return _options[view] || _options;
}
Registry.setComponent('legend.event.helpers.getDisabledOptions', getDisabledOptions);

function getViewByConfig(config, encode) {
    let { ranges, type } = config,
        defaultView = ranges ? LIST_LEGEND_TYPE : encode === 'z' ? NESTED_LEGEND_TYPE : BAND_LEGEND_TYPE;

    return type || defaultView;
}
