// @zohocharts/zc/src/legend/events/legend-events.js
import { LEGEND_EVENT_MAPPINGS } from './legend-events-mappings';
import { TOUCH_EVENTS } from '../../events/eventActions';
import { Registry } from '../../core/zc-registry';
import { LegendEventManager } from './legend-event-manager';
import { d3_event } from '../../core/d3-constants';
import { noOperation } from '../../core/zc-constants';
import { each, getVal, arrayIncludes } from '../../utils/zcutils';
import { triggerUserCallback } from '../../events/customCallbackUtils';
import { getEventObject } from '../../events/events-utils';
import { getHoverOptions } from './legend-event-helpers';
import { $Browser } from '../../core/browser-detect';

let LEGEND_EVENT_NAMESPACE = 'legend';

export class LegendEvents {
    constructor(chart) {
        this.chart = chart;
        this.eventManager = {};
        this.reverseEventManager = {};
        this.init();
    }

    init() {
        let behaviourEvent = this,
            chart = behaviourEvent.chart,
            { legends } = chart;

        legends.forEach(legend => {
            each(LEGEND_EVENT_MAPPINGS, (behaviourMap, name) => {
                behaviourEvent.initiateBehaviour(name, legend);
            });
        });
    }

    update() {
        this.init();
    }

    initiateBehaviour(name, legend) {
        let chart = this.chart,
            { eventManager, reverseEventManager } = this,
            behaviourMap = LEGEND_EVENT_MAPPINGS[name],
            handlers = Registry.getComponent(getVal(behaviourMap._handlers, [chart])),
            { id } = legend,
            isSupported = !handlers._empty && handlers.SUPPORTED(chart, id);

        if (isSupported && legend.enabled) {
            let args = Registry.getComponent(getVal(behaviourMap._arguments, [chart]))(chart, id),
                { BEHAVIOUR, DATA_FINDER, LISTENER, REVERSE_BEHAVIOUR, REVERSE_DATA_FINDER } = handlers,
                managerID = `${name}(${id})`;

            let manager = initManager(managerID, eventManager, chart, args);
            setManager(manager, BEHAVIOUR, DATA_FINDER);
            manager.customBehaviour = getCustomBehaviour(manager);

            if (args.enabled) {
                let { mouseHandlers, touchHandlers } = splitListeners(manager, LISTENER, args.trackerElement),
                    tracker = args.trackerElement;

                manager.bindTo(tracker);
                manager.attachListeners(tracker, mouseHandlers, touchHandlers, null, LEGEND_EVENT_NAMESPACE);
            } else {
                manager.detachListeners(LEGEND_EVENT_NAMESPACE);

                // TODO: check CX_OTHERS_CRMCORE_5
                let hoverOptions = legend.meta.encode === 'clr' && getHoverOptions(legend.meta.config, 'clr');
                if (!(legend.meta.encode === 'clr' && hoverOptions.type && hoverOptions.type !== 'none')) {
                    return;
                }
            }

            let reverseManager = initManager(managerID, reverseEventManager, chart, args);
            setManager(reverseManager, REVERSE_BEHAVIOUR, REVERSE_DATA_FINDER);
        }
    }
}

function initManager(managerID, managerHead, chart, args) {
    managerHead[managerID] = managerHead[managerID] || new LegendEventManager(chart);
    let manager = managerHead[managerID];
    manager.init();
    manager.options = args;

    return manager;
}
function setManager(manager, behaviour, dataFinder) {
    if (!behaviour) {
        return;
    }

    manager.dataFinder(dataFinder);
    manager.defaultBehaviour = behaviour;
    manager.customBehaviour = noOperation;
}

function splitListeners(manager, listeners, group) {
    let mouseHandlers = {},
        touchHandlers = {};

    each(listeners, (listener, type) => {
        let _handler = function(event) {
            manager.targetEvent = d3_event(event);
            listener.apply(manager, [arguments[1], group.nodes().indexOf(this), group, type]);
        };

        if (arrayIncludes(TOUCH_EVENTS, type)) {
            touchHandlers[type] = _handler;
        } else {
            mouseHandlers[type] = _handler;
        }
    });

    return { mouseHandlers, touchHandlers };
}

function getCustomBehaviour(manager) {
    let { options, chart } = manager,
        { events: eventsConfig = {} } = options.config;

    return function(selectionManager, eventType) {
        let tempEventName = $Browser.PHANTOMJS ? { move: 'mousemove', out: 'mouseout' } : { move: 'mouseenter', out: 'mouseleave' };
        let _callbackName =
                eventType === 'dblclick'
                    ? 'doubleclick'
                    : eventType === tempEventName.out
                    ? 'mouseout'
                    : eventType === tempEventName.move
                    ? 'mouseover'
                    : eventType,
            customBehaviour = eventsConfig[_callbackName];

        if (customBehaviour) {
            return [
                {
                    action: triggerUserCallback(customBehaviour, selectionManager.point, getEventObject(manager.targetEvent), null, chart),
                    type: 'custom'
                }
            ];
        }
    };
}

Registry.setComponent('LegendEvents', LegendEvents);
