// @zohocharts/zc/src/pivot/append-nodes.js
import { Branch } from './branches';
import { Cluster, ClusterSet } from './clusters';
import { defined, isUndefined, pick } from '../utils/zcutils';
import { Pane } from './panes';
import { Leaf } from './leaves';
import { ArrayList } from '../utils/array-utils';
import { Node } from './node';
import { createiNode, addBranchToiNode, addLeavesToiNode } from './i-node';

export function pushToBranch(node) {
    let nodeList = this,
        type = node.tag,
        branchIdx = node.getDepth(),
        levelByTag = type == 'row' ? 'levelByRow' : 'levelByColumn',
        levelByTag_invert = type == 'row' ? 'levelByColumn' : 'levelByRow',
        branch;
    (branch = nodeList.branches[branchIdx]) || ((branch = nodeList.branches[branchIdx] = new Branch()) && nodeList.count++);

    let nodeValue = node.mainTag == 'layer' ? '' : node.point || node.value,
        nodeIdx = branch.list.indexOf(nodeValue);

    if (branchIdx != 0 || !defined(nodeIdx)) {
        branch.list.add(node.value);
        branch.index++;
        branch.domain.push(nodeValue || '');
        branch.nodes.push(node);
        let branchNode = createiNode(node);
        addBranchToiNode(node, branchNode);
        branch.inodes.push(branchNode);
        node.index = branch.index;
    } else {
        node.index = nodeIdx;
    }
    node.level = `${pick(node.parent && node.parent.level, '') + node.index}_`;
    node[levelByTag] = `${pick(node.parent && node.parent[levelByTag], '') + node.index}_`;
    node[levelByTag_invert] = `${pick(node.parent && node.parent[levelByTag_invert], '')}`;
}

export function createDuplicateLayerNode(node) {
    let idx = node.tag == 'row' ? node.childIndexbyrow : node.childIndexbycolumn,
        layerName = node.mainTag == 'layer' ? node.value : `manual_${idx}`,
        node_duplicate = new Node(node, node.depthbyrow, node.depthbycolumn, node.depth, 'layer', layerName);
    node_duplicate.node = node.node;
    node_duplicate.row = node.row;
    node_duplicate.column = node.column;
    (node_duplicate.childIndexbyrow = node.childIndexbyrow), (node_duplicate.childIndexbycolumn = node.childIndexbycolumn);
    this.noLayers = node.mainTag !== 'layer';
    return node_duplicate;
}

export function pushToLayers(node, series) {
    let layerList = this,
        rowIdx = node.row,
        columnIdx = node.column,
        layerIdx,
        layer,
        panekey = `${rowIdx}_${columnIdx}`,
        pane,
        panesByCategory,
        hasLayerNode = node.layer;
    layerIdx = node.layer = node.layer && node.layer.length > 1 ? node.layer.indexOf(node.layer[node.childIndexbyrow]) : 0;

    (layer = layerList.groups[layerIdx]) || (layer = layerList.groups[layerIdx] = new ArrayList());
    (pane = layerList.panes[panekey]) || ((pane = layerList.panes[panekey] = new Pane()) && layerList.count++);
    pane.index = layerList.count - 1;
    pane.row = rowIdx;
    pane.column = columnIdx;
    pane.layer = layerIdx;
    pane.data.push(node);
    layerList.domain.add(node.value);
    layer.add(node.value);
    layerList.isLayerNode = hasLayerNode;
    node.levelByColumn = `${pick(node.parent && node.parent.levelByColumn, '')}`;
    node.levelByRow = `${pick(node.parent && node.parent.levelByRow, '')}`;
    // node.level = `${pick(node.parent && node.parent.level, '') + node.index}_`;

    // this.createSeries(node, series);
}

export function createCluster(node) {
    var nodeList = this,
        childIdx = node.childIndex,
        clusters = nodeList.clusters;
    let cluster;
    (cluster = clusters[childIdx]) || ((cluster = clusters[childIdx] = new Cluster()) && nodeList.clustersCount++);

    let nodeIndex = this.clusterMap.indexOf(`${node[this.name]}_${childIdx}`);

    if (isUndefined(nodeIndex)) {
        this.clusterMap.add(`${node[this.name]}_${childIdx}`);
        node.index = this.leafCount++;
    } else {
        node.index = nodeIndex;
    }
}

export function pushToDiscreteCluster(node) {
    if (node.isDefined) {
        let clusters = this.clusters,
            attr = this.levelAttr,
            idx = pick(node.parent && node.parent[attr], ''),
            cluster;
        (cluster = clusters[0]) ||
            ((cluster = clusters[0] = new Cluster()) && (++this.leafCount && ++this.clustersCount) && (cluster.datatype = this.name == 'column' ? node.datatype._columns || 'ordinal' : 'ordinal')); //For ordinal datatype, using single column for for all leaves and also updae time datatype for time data layout
        let clusterSet,
            entries = cluster.entries;
        clusterSet = entries.has(idx) || entries.set(idx, new ClusterSet());
        clusterSet = entries.get(idx);
        let list = clusterSet.list,
            nodeidx = list.indexOf(node.value, true),
            leaf;
        if (isUndefined(nodeidx)) {
            //missing data check
            /*
            To check whether data is missed in any of the grid.
            actualNodeCount - refers total no. of nodes present till previous column (i.e from clsuter.entries[])
            list.count - refers node count for  current column or clusterSet
            cluster.count - refers overall node count in the cluster
             */
            if (!this.missedNodes) {
                let actualNodeCount = cluster.getPrevNodesCountBySetIdx(idx);
                this.missedNodes = actualNodeCount + list.count != cluster.count;
            }

            leaf = createLeaf(node);
            leaf.index = cluster.count;
            nodeidx = list.count;

            clusterSet.domain[nodeidx] = node.value;
            clusterSet.nodes[nodeidx] = leaf;
            clusterSet.list.add(node.value, cluster.count++);
            // cluster.nodes.push(node);
        } else {
            clusterSet.nodes[nodeidx].data.push(node.data);
        }
        var inode = this.branches[node.depthbycolumn] ? this.branches[node.depthbycolumn].inodes[node.column] : null;
        if (inode) {
            addLeavesToiNode(node, this.branches[node.depthbycolumn].inodes[node.column], clusterSet.nodes[nodeidx]);
        }

        if (isUndefined(cluster.list.indexOf(node.value))) {
            cluster.list.add(node.value);
            cluster.keys.push(node.value);
        }
        node.data.index = node.index = list.indexOf(node.value);
        node.data.level = `${pick(node.parent && node.parent.level, '') + node.index}_`;
        node.data.levelByColumn = `${pick(node.parent && node.parent.levelByColumn, '') + node.index}_`;
        node.data.levelByRow = `${pick(node.parent && node.parent.levelByRow, '') + node.index}_`;

        if (leaf) {
            leaf.level = node.data.level;
            leaf.levelByColumn = node.data.levelByColumn;
            leaf.levelByRow = node.data.levelByRow;
        }
    }
}

function createLeaf(node) {
    let leaf = new Leaf(node.value, node.depth, node.depthbyrow, node.depthbycolumn);
    leaf.data = [node.data];
    leaf.node = node;
    leaf.parent = node.parent;
    return leaf;
}

export function pushToContinuousCluster(node) {
    let clusters = this.clusters,
        idx = this.name == 'row' ? node.childIndexbyrow : node.childIndexbycolumn,
        cluster;
    (cluster = clusters[idx]) || ((cluster = clusters[idx] = new Cluster()) && ++this.leafCount && ++this.clustersCount);

    if(this.name == 'column'){
        node.data.level = `${pick(node.parent && node.parent.level, '') + node.index}_`;
        node.data.levelByColumn = `${pick(node.parent && node.parent.levelByColumn, '') + node.index}_`;
        node.data.levelByRow = `${pick(node.parent && node.parent.levelByRow, '') + node.index}_`;
    }
    cluster.nodes.push(node);
    // cluster.domain.push(node.value);
    clusters.count++;
}
