// @zohocharts/zc/src/pivot/categories-range.js
import { defined, pick } from '../utils/zcutils';
import { d3_map, d3_scaleOrdinal, d3_range } from '../core/d3-constants';
import { mathAbs } from '../core/zc-constants';

export function setRanges(tag, leafscaleUtil) {
    let mode = leafscaleUtil ? 'discrete' : 'continuous',
        nodeList = tag == 'row' ? this.rows : this.columns,
        leafscale,
        pointProperty,
        startRange,
        sign_correction = 1,
        splitLeavesColumn = tag == 'column' ? nodeList.splitColumnLeaves : null,
        uniformLeafScale;

    if (mode == 'discrete') {
        leafscaleUtil._padding(nodeList.getClusterNodes(0), nodeList.getDomainForCluster(0));
        nodeList.leafScale = leafscale = leafscaleUtil._scale; //TODO check whether this property need tor not
        if (splitLeavesColumn) {
            var rangeInfo = leafscaleUtil._rangeInfo();
            startRange = rangeInfo.startPos;
            sign_correction = rangeInfo.sign;
            uniformLeafScale = leafscaleUtil._haUniformLeafScale;
        }
        pointProperty = leafscaleUtil.pointProperty;
    }
    var count = nodeList.count;
    for (var i = 0; i < count; i++) {
        var domain = nodeList.branches[i].nodes,
            range = (nodeList.branches[i].range = new Array(domain.length)),
            inodes = nodeList.branches[i].inodes,
            endRange = startRange;
        domain.forEach((d, i) => {
            if (d.children) {
                if (isFirstBranch(tag, d.children)) {
                    var c = inodes[i].children,
                        children = c && c.length ? c : d.children;

                    d.rangeBand = getRangeBandForFirstBranch(nodeList, children, mode, i);
                    var tempRange = mode == 'continuous' ? d.rangeBand * i : splitLeavesColumn ? endRange : getStartRange(children);
                    endRange = tempRange + sign_correction * d.rangeBand;
                    d.range = [tempRange, endRange];
                } else {
                    d.rangeBand = d.children.reduce(sumLeaves, 0);
                    var firstProperChild = getFirstFullData(d.children);
                    d.range = firstProperChild ? [firstProperChild.range[0], d.rangeBand + firstProperChild.range[0]] : [];
                }
                range[i] = d.range;
            }
        });
    }

    function getStartRange(children) {
        let startRange;
        if (leafscale) {
            let firstDefinedData = isLeafChild(children) ? getFirstDefinedData(children) : getFirstDefinedData(children[0].children);
            startRange = leafscale(firstDefinedData[pointProperty]) - getSpan(firstDefinedData, true, true);
        }
        return startRange;
    }
    function getRangeBandForFirstBranch(nodeList, children, mode, childIndex) {
        if (mode == 'continuous' || splitLeavesColumn) {
            return getContinousRangeBandForFirstBranch(nodeList, childIndex);
        } else {
            return getDiscreteRangeBandForFirstBranch(nodeList, children);
        }
    }

    function getContinousRangeBandForFirstBranch(nodeList, childIndex) {
        if (!uniformLeafScale && nodeList.nonUniformLeafSize) {
            return nodeList.nonUniformLeafSize[childIndex] * nodeList.clustersCount;
        }
        return nodeList.leafSize * nodeList.clustersCount;
    }

    function getDiscreteRangeBandForFirstBranch(nodeList, children) {
        if (leafscaleUtil._rangebandForFirstBranch) {
            return leafscaleUtil._rangebandForFirstBranch(children);
        }
        if (isLeafChild(children)) {
            return children.reduce(sumLeaves, 0);
        }
        let mergedChildren = mergerLayerChildren(children);
        if (mode == 'discrete' && leafscaleUtil._hasDynamicLeafScale()) {
            leafscaleUtil._padding(nodeList.getClusterNodes(0), nodeList.getDomainForCluster(0), mergedChildren[0].index);
        }
        return mergedChildren.reduce(sumLeaves, 0);
    }

    function mergerLayerChildren(layers) {
        let children = d3_map();
        layers.map(d => {
            d.children.map(child => {
                if (!children.has(child.value)) {
                    var node = child.rootLayer ? child : child.node || child;
                    children.set(child.value, node);
                }
            });
        });
        return children.values();
    }

    function isLeafChild(children) {
        return defined(children[0]) && children[0].tag === 'leaf';
    }
    function isFirstBranch(tag, children) {
        return (
            children && defined(children[0]) && (children[0].tag != tag || children[0].tag === 'leaf' || children[0].datatype == 'numeric')
        );
    }

    function sumLeaves(a, point) {
        if (!point.isDefined || point.isEmpty) {
            return 0;
        }
        if (!point.rangeBand) {
            return a + leafscale.bandwidth(point.index) + getSpan(point, true, leafscale) + getSpan(point, false, leafscale);
        }
        return a + point.rangeBand;
    }

    function getSpan(point, startSide, useStaticPadding) {
        var node = point.name == 'Leaf' ? point.node : point;
        if ((startSide && node === leafscale.firstPt) || (!startSide && node === leafscale.lastPt)) {
            return useStaticPadding ? pick(leafscale._outerPadding, leafscale.outerPadding) : leafscale.outerPadding;
        } else {
            return useStaticPadding ? pick(leafscale._innerPadding, leafscale.innerPadding) : leafscale.innerPadding;
        }
    }

    function getFirstDefinedData(children) {
        var firstDefinedChild;
        children.some(d => {
            var node = d.name == 'Leaf' ? d.node : d;
            if (node.isDefined) {
                firstDefinedChild = d;
            }
            return node.isDefined;
        });
        return firstDefinedChild;
    }

    function getFirstFullData(children) {
        var firstDefinedChild;
        children.some(d => {
            if (!d.isEmpty) {
                firstDefinedChild = d;
            }
            return !d.isEmpty;
        });
        return firstDefinedChild;
    }
}

/*
Ranges for non linear ordinal leaf scale
 */
export function setRangesForLeafScale(tag, leafScale) {
    var nodeList = tag == 'row' ? this.rows : this.columns,
        actualDomain = leafScale.domain(),
        bandwidth = leafScale.bandwidth(),
        categoriesCount_firstBranch = nodeList.count ? nodeList.branches[0].domain.length : 1,
        clusterSet = nodeList.clusters[0];
    var counter = 0,
        counter1 = 0;
    var range = new Array(0);
    for (var i = 0; i < categoriesCount_firstBranch; i++) {
        var dynamicDomain = clusterSet.entries.values()[i].customDomain,
            domain = nodeList.getDomainForClusterSet(0, i);
        domain.map((d, j) => {
            var startCounter = counter,
                endCounter,
                tempDynamicScale;

            if (dynamicDomain && dynamicDomain.get(0) && j == domain.length - 1) {
                //handled only for single data as children for category

                var noOfDummyCategories = dynamicDomain.get(0).count;
                // if (domain.length > 1) {
                //     if (!tempDynamicScale) {
                //         tempDynamicScale = d3_scaleOrdinal()
                //             .domain(d3_range(domain.length))
                //             .range([leafScale(startCounter), leafScale(startCounter + domain.length + noOfDummyCategories - 1)])
                //             .paddingOuter(0)
                //             .paddingInner(leafScale.paddingInner());
                //     }
                //     // var x = mathAbs(leafScale(0) - leafScale(noOfDummyCategories - 1)) / noOfDummyCategories;
                //     // range[counter1] = [leafScale(startCounter), leafScale(startCounter) + bandwidth + x];
                //
                //     range[counter1] = [tempDynamicScale(startCounter), tempDynamicScale(startCounter) + tempDynamicScale.bandwidth()];
                //     if (j == domain.length - 1) {
                //         counter = startCounter + noOfDummyCategories;
                //     }
                // } else {
                endCounter = startCounter + noOfDummyCategories;
                range[counter1] = [leafScale(startCounter), leafScale(endCounter) + bandwidth];
                counter = endCounter;
                // }
            } else {
                range[counter1] = [leafScale(startCounter), bandwidth + leafScale(startCounter)];
            }

            counter++;
            counter1++;
        });
    }
    return range;
}
