// @zohocharts/zc/src/pivot/clusters.js
import { ArrayList } from '../utils/array-utils';
import { d3_map } from '../core/d3-constants';

export function Cluster() {
    this.datatype = 'numeric';
    this.nodes = new Array(0);
    this.domain = new Array(0);
    this.range = [];
    this.entries = d3_map();
    this.keys = [];
    this.index = 0;
    this.count = 0;
    this.list = new ArrayList();
}

/*
To get the total no. of nodes in cluster.entries up to the given idx
 */
function getPrevNodesCountBySetIdx(idx) {
    if (this.entries.get(idx).prevNodesCount) {
        return this.entries.get(idx).prevNodesCount;
    }
    let nodeCount = 0;
    for (var i = 0; i < idx; i++) {
        nodeCount += this.entries.has(i) ? this.entries.get(i).list.count : 0;
    }
    this.entries.get(idx).prevNodesCount = nodeCount;
    return nodeCount;
}

Cluster.prototype = {
    constructor: Cluster,
    getPrevNodesCountBySetIdx: getPrevNodesCountBySetIdx
};

/*
clusterSet is subset of cluster. It is stored in cluster.entries
 */
export function ClusterSet() {
    this.list = new ArrayList();
    this.domain = new Array(0);
    this.nodes = new Array(0);
    this.prevNodesCount = null;
}

ClusterSet.prototype = {
    constructor: ClusterSet
};
