// @zohocharts/zc/src/pivot/columns.js
import { pushToBranch, createCluster, pushToDiscreteCluster, pushToContinuousCluster } from './append-nodes';
import { ArrayList } from '../utils/array-utils';
import {
    getNodelistCount,
    getBranchesCount,
    getLeavesCount,
    getClustersCount,
    getDomainForBranch,
    getRangesForBranch,
    getAllChildren,
    hasMultiCluster,
    getDomainForCluster,
    getClusterNodes,
    getNodesAtDepth,
    getBranchNodes,
    getUniqueDomainForCluster,
    setNodeIndexForClusters,
    getDomainForClusterSet,
    getNodesForClusterSet,
    getDynamicDomainForClusterSet,
    getMergedDomainForClusterSet,
    getDynamicDomainForCluster
} from './nodelist-helpers';
import { getParentByIndex } from './parent';

export function Column() {
    this.count = 0;
    this.clustersCount = 0;
    this.leafCount = 0;
    this.branches = [];
    this.clusters = [];
    this.name = 'column';
    this.levelAttr = 'levelByColumn';
    this.leafNodeMap = new ArrayList();
}

Column.prototype = {
    constructor: Column,
    pushToBranch: pushToBranch,
    createCluster: createCluster,
    pushToDiscreteCluster: pushToDiscreteCluster,
    pushToContinuousCluster: pushToContinuousCluster,
    getCount: getNodelistCount,
    getBranchesCount: getBranchesCount,
    getLeavesCount: getLeavesCount,
    getClustersCount: getClustersCount,
    hasMultiCluster: hasMultiCluster,
    getDomainForBranch: getDomainForBranch,
    getRangesForBranch: getRangesForBranch,
    getDomainForCluster: getDomainForCluster,
    getUniqueDomainForCluster: getUniqueDomainForCluster,
    getDomainForClusterSet: getDomainForClusterSet,
    getNodesForClusterSet: getNodesForClusterSet,
    getDynamicDomainForClusterSet: getDynamicDomainForClusterSet,
    getMergedDomainForClusterSet: getMergedDomainForClusterSet,
    getDynamicDomainForCluster: getDynamicDomainForCluster,
    getClusterNodes: getClusterNodes,
    getBranchNodes: getBranchNodes,
    getNodesAtDepth: getNodesAtDepth,
    getParent: getParentByIndex,
    getAllChildren: getAllChildren,
    setNodeIndexForClusters: setNodeIndexForClusters
};
