// @zohocharts/zc/src/pivot/i-node.js
import { ArrayList } from '../utils/array-utils';

export function iNode(tree, name) {
    this.value = name;
    this.children = new Array(0);
    this.layerNames = new ArrayList();
    this.leafNames = new ArrayList();
    this.list = new ArrayList();
    this.node = tree;
}

export function createiNode(node) {
    let i_node = new iNode(node, node.value);
    return i_node;
}

export function addBranchToiNode(node, inode) {
    let tag = node.tag,
        isSingleBranch = tag == 'row' ? node.firstRow && node.lastRow : node.firstColumn && node.lastColumn,
        isFirstBranch = tag == 'row' ? node.firstRow : node.firstColumn,
        isLastBranch = tag == 'row' ? node.lastRow : node.lastColumn;
    if (!isLastBranch) {
        inode.children = node.children;
    }
}

export function addLeavesToiNode(node, inode, leaf) {
    if (!node.rootLayer) {
        var parentName = node.parent.value,
            leafMapKey = `${parentName}_${leaf.value}`;
        if (node.parent.tag == 'layer') {
            var layerNode;
            if (inode.layerNames.indexOf(parentName) == null) {
                layerNode = createiNode(node.parent);
                inode.layerNames.add(parentName);
                layerNode.children.push(leaf);
                inode.children.push(layerNode);
            } else {
                if (inode.leafNames.indexOf(leafMapKey) == null) {
                    layerNode = inode.children[inode.layerNames.indexOf(parentName)];
                    layerNode.children.push(leaf);
                }
            }
        } else {
            leafMapKey = `${leaf.value}`;
            if (inode.leafNames.indexOf(leafMapKey) == null) {
                inode.children.push(node);
            }
        }
        inode.leafNames.add(leafMapKey);
    }
}
