// @zohocharts/zc/src/pivot/index.js
import { Node } from './node';
import { eachNode, eachNodeBeforeAfter } from './node-iterator';
import { Row } from './rows';
import { Column } from './columns';
import { Layer } from './layers';
import { processChild } from './children';
import { setRanges, setRangesForLeafScale } from './categories-range';
import { setLeafSize } from './leaf-size';
import { Registry } from '../core/zc-registry';
import { deleteAllEmptyParents } from './remove-nodes';

export function PivotLayout(data, hooks) {
    let _this = this,
        root = new Node(data, -1, -1, -1, 'root', ''),
        node,
        child,
        childs,
        nodes = [root],
        i,
        n,
        children = hooks.children || defaultChildren,
        nodetype = hooks.nodeType || defaultNodeType,
        datatype = hooks.dataType || defaultDataType,
        name = hooks.nodeName || defaultNodeName,
        nodedata = hooks.data,
        isDefined = hooks.isDefined,
        isEmpty = hooks.isEmpty,
        splitColumnLeaves = hooks.hasSplitLeavesColumns();

    (this.rows = new Row()), (this.columns = new Column()), (this.layers = new Layer()), (this.series = []);

    eachNode(root, node => {
        if (node.tag !== 'leaf' && (childs = children(node)) && (n = childs.length)) {
            node.children = new Array(0);
            var isAllChildEmpty = true;
            for (i = 0; i < n; i++) {
                child = new Node(childs[i]);
                child.parent = node;
                child.depth = node.depth + 1;
                child.row = node.row || 0;
                child.column = node.column || 0;
                child.tag = nodetype(child);
                child.datatype = datatype(child);
                child.depthbyrow = child.tag == 'row' ? node.depthbyrow + 1 : node.depthbyrow;
                child.depthbycolumn = child.tag == 'column' ? node.depthbycolumn + 1 : node.depthbycolumn;
                child.childIndex = node.children.length;
                child.childIndexbyrow = node.childIndexbyrow || 0;
                child.childIndexbycolumn = node.childIndexbycolumn || 0;
                // child.level = `${pick(node.level, '') + i}_`;
                child.isDefined = child.tag == 'leaf' ? isDefined(child.node) : true;
                child.value = name(child.node);
                child.data = nodedata(child.node);
                child.isEmpty = child.tag == 'leaf' ? !child.isDefined : isEmpty(child);
                isAllChildEmpty = isAllChildEmpty ? child.isEmpty : isAllChildEmpty;

                if (!child.isEmpty) {
                    node.children.push(child);
                }
                // this.process(child);
            }
            node.isEmpty = isAllChildEmpty;
            if (node.isEmpty) {
                deleteAllEmptyParents(node);
            }
        }
    });

    this.columns.splitColumnLeaves = splitColumnLeaves;

    eachNodeBeforeAfter(
        root,
        function(node) {
            _this.process(node);
        },
        this.postProcess
    );

    this.rows.branches.reverse();
    this.columns.branches.reverse();

    if (this.columns.missedNodes) {
        this.columns.setNodeIndexForClusters();
    }
}

function postProcess(node, child) {
    let tag = node.tag;
    child.row = node.row || 0;
    child.column = node.column || 0;
    if (tag == 'row') {
        child.childIndexbyrow = node.childIndexbyrow;
    } else if (tag == 'column') {
        child.childIndexbycolumn = node.childIndexbycolumn;
    }
}

function defaultChildren(d) {
    return d.children;
}

function defaultNodeType(d) {}

function defaultDataType(d) {}

function defaultNodeName(d) {}

function getRootNodeList(tag) {
    if (tag == 'row') {
        return this.rows;
    } else if (tag == 'column') {
        return this.columns;
    } else if (tag == 'layer') {
        return this.layers;
    }
}

PivotLayout.prototype = {
    row: Row,
    column: Column,
    layer: Layer,
    process: processChild,
    getRootNodeList: getRootNodeList,
    setRanges: setRanges,
    setRangesForLeaf: setRangesForLeafScale,
    leafSize: setLeafSize,
    postProcess: postProcess
};

Registry.setComponent('PivotLayout', PivotLayout);
