// @zohocharts/zc/src/pivot/layer-hierarchy-execute.js
import { LayerHierarchy } from './layer-hierarchy-index';
import { eachNodeBeforeAfter } from './node-iterator';
import { processChild } from './children';
import { Row } from './rows';
import { Column } from './columns';
import { Layer } from './layers';
import { setRanges } from './categories-range';
import { setLeafSize } from './leaf-size';
import { pick } from '../utils/zcutils';
import { Registry } from '../core/zc-registry';

export function LayerHierarchyLayout(data, hooks) {
    let root,
        _this = this;

    (this.rows = new Row()), (this.columns = new Column()), (this.layers = new Layer());
    this.series = [];

    data.forEach((d, i) => {
        /*
            Added Domain names initially to have missed series data
         */
        this.layers.createSeries({ node: d }, this.series);
        this.layers.domain.add(d.seriesname);
        /*
         */
        if (!d.disabled) {
            root = new LayerHierarchy(d, root, hooks);
        }
    });

    this.columns.splitColumnLeaves = hooks.hasSplitLeavesColumns();

    eachNodeBeforeAfter(
        root,
        function(node) {
            if (node.tag == 'layer') {
                node.data = [
                    node.children.map(d => {
                        return d.data;
                    })
                ];
            }
            _this.process(node);
        },
        this.postProcess
    );

    this.rows.branches.reverse();
    this.columns.branches.reverse();
}

function preProcess(node) {
    if (node.tag == 'layer') {
        node.data = [
            node.children.map(d => {
                return d.data;
            })
        ];
    }
    this.process(node);
}

function postProcess(node, child) {
    child.row = node.row || 0;
    child.column = node.column || 0;
}

function getRootNodeList(tag) {
    if (tag == 'row') {
        return this.rows;
    } else if (tag == 'column') {
        return this.columns;
    } else if (tag == 'layer') {
        return this.layers;
    }
}

LayerHierarchyLayout.prototype = {
    row: Row,
    column: Column,
    layer: Layer,
    process: processChild,
    preProcess: preProcess,
    postProcess: postProcess,
    getRootNodeList: getRootNodeList,
    setRanges: setRanges,
    leafSize: setLeafSize,
    executor: LayerHierarchy
};

Registry.setComponent('LayerHierarchyLayout', LayerHierarchyLayout);
