// @zohocharts/zc/src/pivot/nodelist-helpers.js
import { defined, pick, objectValues } from '../utils/zcutils';
import { d3_merge } from '../core/d3-constants';
import { d3_values } from '../utils/zcutils';

export function getNodelistCount() {
    return this.count;
}

export function getBranchesCount() {
    return this.branches.length;
}

export function getLeavesCount() {
    if (this.name == 'column' && !this.splitColumnLeaves) {
        return this.leafCount;
    }
    if (this.count) {
        return this.clustersCount * this.getDomainForBranch(0).length;
    }
    return this.leafCount;
}

export function getClustersCount() {
    return this.clustersCount;
}

export function hasMultiCluster() {
    return this.clustersCount > 1;
}

export function getLayersCount() {
    return this.count;
}

export function getPanesCount() {
    return this.getPanes().length;
}

export function getDomainForBranch(index) {
    let branch = this.branches[index];
    if (branch) {
        return branch.domain;
    }
}

export function getRangesForBranch(index) {
    let branch = this.branches[index];
    if (branch) {
        return branch.range;
    }
}

export function getDomainForCluster(index) {
    let cluster = this.clusters[index],
        entries = cluster.entries;
    if (cluster) {
        if (entries.size()) {
            return d3_merge(
                entries.values().map(d => {
                    return d.domain;
                })
            );
        } else {
            return cluster.domain;
        }
    }
}

export function getDomainForClusterSet(index, subIndex) {
    let cluster = this.clusters[index];
    if (cluster) {
        return cluster.entries.values()[subIndex].domain;
    }
}

export function getNodesForClusterSet(index, subIndex) {
    let cluster = this.clusters[index];
    if (cluster) {
        return cluster.entries.values()[subIndex].nodes;
    }
}

export function getDynamicDomainForCluster(index) {
    let nodelist = this,
        cluster = this.clusters[index],
        entries = cluster.entries;
    if (cluster) {
        if (entries.size()) {
            return d3_merge(
                entries.values().map((d, i) => {
                    return nodelist.getDynamicDomainForClusterSet(index, i);
                })
            );
        } else {
            return cluster.domain;
        }
    }
}

export function getDynamicDomainForClusterSet(index, subIndex) {
    let cluster = this.clusters[index];
    if (cluster) {
        var customDomain = cluster.entries.values()[subIndex].customDomain;
        if (customDomain) {
            var customDomain_values;
            customDomain.values().map(d => {
                customDomain_values = d.values;
            });
            return d3_merge([cluster.entries.values()[subIndex].domain, customDomain_values]);
        } else {
            return cluster.entries.values()[subIndex].domain;
        }
    }
}

export function getMergedDomainForClusterSet(index, subIndex) {
    let cluster = this.clusters[index];
    if (cluster) {
        var domain = this.getDomainForClusterSet(index, subIndex),
            customDomain = this.getDynamicDomainForCluster(index, subIndex);
        return d3_merge([domain, customDomain]);
    }
}

export function getUniqueDomainForCluster(index) {
    let cluster = this.clusters[index];
    if (cluster) {
        return cluster.keys;
    }
}
export function getBranchNodes(index) {
    let branch = this.branches[index];
    if (branch) {
        return branch.nodes;
    }
}

export function getClusterNodes(index) {
    let cluster = this.clusters[index],
        entries = cluster.entries;
    if (cluster) {
        if (entries.size()) {
            return d3_merge(
                entries.values().map(d => {
                    return d.nodes;
                })
            );
        } else {
            return cluster.nodes;
        }
    }
}

export function getLayerNames() {
    return this.domain.list;
}

export function getParent(node) {
    return node.parent;
}

export function getPanes() {
    return objectValues(this.panes);
}

/**
 *
 * @param {Object} parent
 * @param {Array|*} array for recursion, need not to pass any
 * @returns {Array}
 */
export function getAllChildren(parent, array) {
    array = pick(array, []);
    if (parent.children) {
        parent.children.forEach(function(child) {
            if (child.tag == 'leaf' && child.isDefined) {
                array.push(child.data); // push to existing array
            } else {
                getAllChildren(child, array);
            }
        });
    } else {
        return parent.data;
    }
    return array;
}

export function getNodesAtDepth(order) {
    //order refers axisorder
    let leavesCount = this.getLeavesCount();
    if (order > leavesCount - 1) {
        return this.getBranchNodes(order - leavesCount);
    } else {
        return this.getClusterNodes(0);
    }
}

export function setNodeIndexForClusters() {
    let nodeList = this;
    nodeList.clusters.map((cluster, i) => {
        let clusterNodes = nodeList.getClusterNodes(i);
        clusterNodes.map((node, i) => {
            node.index = i;
            node.data.map(d => {
                //Remove the old index value and replaced with new index
                d.index = i;
                node.level = d.level = d.level
                    .split('_')
                    .slice(0, -2)
                    .concat([d.index])
                    .join('_');
                node.levelByColumn = d.levelByColumn = d.levelByColumn
                    .split('_')
                    .slice(0, -2)
                    .concat([d.index])
                    .join('_');
                node.levelByRow = d.levelByRow = d.levelByRow
                    .split('_')
                    .slice(0, -2)
                    .concat([d.index])
                    .join('_');
                // d.level = `${pick(node.parent && node.parent.levelByColumn, '') + d.index}_`;
                // d.levelByColumn = `${pick(node.parent && node.parent.levelByColumn, '') + d.index}_`;
                // d.levelByRow = `${pick(node.parent && node.parent.levelByRow, '') + d.index}_`;
            });
        });
    });
}

export function getPaneIndex(grid) {
    return `pane${grid.index || 0}`;
}
