// @zohocharts/zc/src/pivot/rows.js
import { pushToBranch, createCluster, pushToDiscreteCluster, pushToContinuousCluster } from './append-nodes';
import { ArrayList } from '../utils/array-utils';
import {
    getNodelistCount,
    getBranchesCount,
    getLeavesCount,
    getClustersCount,
    getDomainForBranch,
    getRangesForBranch,
    getAllChildren,
    hasMultiCluster,
    getDomainForCluster,
    getClusterNodes,
    getNodesAtDepth,
    getBranchNodes,
    getDomainForClusterSet,
    getNodesForClusterSet
} from './nodelist-helpers';
import { getParentByIndex } from './parent';

export function Row() {
    this.count = 0;
    this.clustersCount = 0;
    this.leafCount = 0;
    this.branches = [];
    this.clusters = [];
    this.name = 'row';
    this.levelAttr = 'levelByRow';
    this.clusterMap = new ArrayList();
}

Row.prototype = {
    constructor: Row,
    pushToBranch: pushToBranch,
    createCluster: createCluster,
    pushToDiscreteCluster: pushToDiscreteCluster,
    pushToContinuousCluster: pushToContinuousCluster,
    getCount: getNodelistCount,
    getBranchesCount: getBranchesCount,
    getLeavesCount: getLeavesCount,
    getClustersCount: getClustersCount,
    hasMultiCluster: hasMultiCluster,
    getDomainForBranch: getDomainForBranch,
    getRangesForBranch: getRangesForBranch,
    getDomainForCluster: getDomainForCluster,
    getDomainForClusterSet: getDomainForClusterSet,
    getNodesForClusterSet: getNodesForClusterSet,
    getClusterNodes: getClusterNodes,
    getBranchNodes: getBranchNodes,
    getNodesAtDepth: getNodesAtDepth,
    getParent: getParentByIndex,
    getAllChildren: getAllChildren
};
