// @zohocharts/zc/src/pivot/series.js
export function createSeries(node, series) {
    let layers = this,
        data = node.node,
        domain = layers.splitClusters.count ? layers.splitClusters : layers.domain;

    if (!series[domain.indexOf(node.value)]) {
        var obj = {
            type: data.type,
            color: data.color,
            data: [[]],
            yaxiscolumnorder: data.yaxiscolumnorder,
            seriesname: data.seriesname,
            disabled: data.disabled,
            hideInLegend: data.hideInLegend,
            filtered: data.filtered
        };
        if (data.plotoptions) {
            obj.plotoptions = data.plotoptions;
        }

        series.push(obj);
    }
}

export function pushToSeries(node, series) {
    let layers = this,
        domain = layers.splitClusters.count ? layers.splitClusters : layers.domain;
    var seriesIdx = domain.indexOf(getSeriesName(node, layers.splitClusters.count)) || 0;
    if (series[seriesIdx]) {
        series[seriesIdx].data[0].push(node.data);
    }
}

function getSeriesName(node, splitClusters) {
    let layername = node.seriesname;
    if (!layername) {
        if (node.parent.tag == 'layer') {
            layername = splitClusters ? node.parent.parent.value : node.parent.value;
        } else if (node.parent.datatype == 'numeric') {
            layername = node.parent.value;
        } else {
            let idx = node.parent.tag == 'row' ? node.childIndexbyrow : node.childIndexbycolumn;
            layername = `manual_${idx}`;
        }
    }
    return layername;
}
