// @zohocharts/zc/src/axis/cartesian/mark-helpers/axistitle-prop.js
import { pick } from '../../../utils/zcutils';
import { getAxisAreaDim } from '../../hierarchy-axis-utils';
import { AxisRegistry } from '../../axis-registry';
import { getFormattedValue } from '../../../formatter/formatter';
import { applyFont } from '../../../utils/zcsvgutils';
import { $ZCG } from '../../../core/zcglobal';
import { mathAbs } from '../../../core/zc-constants';
import { d3_min } from '../../../core/d3-constants';

export function getAxisTitlePosition() {
    let _ = this,
        { axisTicks } = _,
        { tickInfo } = axisTicks,
        { axisgroup } = axisTicks.precalculate,
        { axisTitle } = _;
    var rotations = {
            top: 0,
            right: 90,
            bottom: 0,
            left: 270
        },
        textAnchor = 'middle', // it is middle for all angle
        rotate = rotations[_.orient],
        x,
        y,
        dy = '-0.25em',
        axisRect = axisgroup,
        plotBound = axisTitle.precalculate ? getAxisAreaDim(_.type, _.order, _.orient, _.chart, _.chart.plotarea) : null,
        plotRect = pick(plotBound, this.plotRect),
        tickLabelConfig = _.config.ticklabel,
        axisLabelConfig = axisTitle.branchTitleConfig,
        tickMargin = axisTicks.margin,
        labelMargin = axisTitle.titleMargin;
    var formattedValue = getFormattedValue({}, axisLabelConfig.label.text, 'ordinal', this.chart, 'axis'),
        // fontSize = parseFloat(utils.getVal(getValByPriority([label.config, systemConfig.chart, systemConfig.canvas, $ZCG], 'fontSize'))),
        styleObj = applyFont(null, [axisLabelConfig.label, this.chart.systemConf.chart, this.chart.systemConf.canvas, $ZCG]);
    //
    // // reset BBox
    // if (label.fontSize !== fontSize) {
    //     label.rect = null;
    // } else {
    //     label.fontSize = fontSize;
    // }

    // if axis rotated is called;
    // label.element.attr('dy', '-0.25em');
    var isInline = tickInfo.isInline,
        isInwardToPlot = tickInfo.inlineToPlot,
        isOutwardToPlot = isInline && !isInwardToPlot,
        offset = isOutwardToPlot ? 0 : tickMargin[_.orient],
        _attrs;

    var { axisInfo } = _,
    scale = axisInfo.scale || _.axisTicks.tickInfo.scale,
    range = scale.origRange ? scale.origRange : scale.range(),
    rangeDiff = mathAbs(range[1] - range[0]);

    switch (_.orient) {
        case 'top': {
            var tickLength = isInwardToPlot ? 0 : axisRect.height + offset;
            x = plotRect.width / 2;
            y = tickLength + labelMargin.bottom;
            _attrs = getParsedAttrs(rotate, x, -y);
            break;
        }

        case 'right': {
            // label.element.attr("dy", "-0.25em");
            var tickLength = isInwardToPlot ? 0 : axisRect.width + offset;
            x = plotRect.height / 2;
            y = tickLength + labelMargin.left;
            _attrs = getParsedAttrs(rotate, x, -y);
            break;
        }
        case 'bottom': {
            var tickLength = isInwardToPlot ? 0 : axisRect.height + offset;
            x = _.chart.defaultScroll ? plotRect.width / 2 : d3_min([range[0], range[1]]) + rangeDiff / 2;
            y = tickLength + labelMargin.top;

            // apply baseline
            dy = '0.91em';
            // label.element.attr('dy', '0.91em');
            _attrs = getParsedAttrs(rotate, x, y);
            break;
        }
        case 'left': {
            // label.element.attr("dy", "-0.25em");
            var tickLength = isInwardToPlot ? 0 : axisRect.width + offset;
            x = tickLength + labelMargin.right;
            y = plotRect.height / 2;
            _attrs = getParsedAttrs(rotate, -x, y, -x, y);
            break;
        }
    }

    let titleProp = {
        x: _attrs.x,
        y: _attrs.y,
        transform: _attrs.transform,
        dy,
        textAnchor,
        text: formattedValue,
        styleObj
    };

    return titleProp;
}

AxisRegistry.add('cartesian.axisTitle', getAxisTitlePosition);

function getParsedAttrs(rotate, x, y, x1, y1) {
    return {
        x,
        y,
        transform: `rotate(${rotate} ${x1 || 0} ${y1 || 0})`
    };
}
