// @zohocharts/zc/src/axis/cartesian/mark-helpers/axisline-prop.js
import { AxisRegistry } from '../../axis-registry';

export function getAxisLinePosition() {
    let _ = this,
        attrs = {
            x1: 0,
            y1: 0,
            x2: _.orient == 'bottom' || _.orient == 'top' ? _.plotRect.width : 0.0001, // phantomjs 1.9.8 issue
            y2: _.orient == 'bottom' || _.orient == 'top' ? 0.0001 : _.plotRect.height // // phantomjs 1.9.8 issue
        };
    return attrs;
}

export function getAxisLinePathVal(positionH) {
    let _ = this,
        { axisInfo } = _,
        scale = axisInfo.scale || _.axisTicks.tickInfo.scale,
        range = scale.origRange ? scale.origRange : scale.range(),
        range0 = +range[0] + 0.5,
        range1 = +range[range.length - 1] + 0.5,
        tickSizeOuter = 0,
        d =
            _.orient === 'left' || _.orient == 'right'
                ? `M${positionH.attrs_value.k * (tickSizeOuter + positionH.tickmarkStart)},${range0}${'h'}0.5V${range1}H${positionH
                      .attrs_value.k *
                      (tickSizeOuter + positionH.tickmarkStart)}`
                : `M${range0},${positionH.attrs_value.k * tickSizeOuter}V0.5H${range1}V${positionH.attrs_value.k * tickSizeOuter}`;
    return d;
}

AxisRegistry.add('cartesian.axisLinePosition', getAxisLinePosition);
AxisRegistry.add('cartesian.axisLinePath', getAxisLinePathVal);
