// @zohocharts/zc/src/axis/cartesian/mark-helpers/ticklabel-prop.js
import { AxisRegistry } from '../../axis-registry';
import { invokeFunction } from '../../../utils/zcutils';

export function getTicklabelPosition(positionH) {
    let _ = this,
        { tickInfo } = _.axisTicks,
        isLevelType = this.axisTicks.tickH.isHierarchialLevelType(),
        attrs = {
            [positionH.attrs_key.x]: positionH.l * positionH.spacing,
            [positionH.attrs_key.y]: 0,
            dy:
                tickInfo && tickInfo.mode == 'zigzag' && positionH.rotationInfo.dy1
                    ? positionH.rotationInfo.dy1
                    : positionH.rotationInfo.dy, //(In prerender used default as dy1 for zigzag used in axis-prerender-marks.js)
            _dy: positionH.rotationInfo.dy, //default dy (used in actual rendering of ticklabel axis-marks.js)
            transform: `rotate(${positionH.rotationInfo.rotate[0]} ${positionH.rotationInfo.rotate[1]} ${
                positionH.rotationInfo.rotate[2]
            })`,
            //transform -> prerender
            //transform1 -> render
            transform1: function(d, i) {
                let scale = _.axisInfo.scale || tickInfo.scale,
                    t = [0, 0];
                if (_.config.tickTransform) {
                    t = invokeFunction(_.config.tickTransform, d, scale, _.chart);
                }
                if (isLevelType) {
                    /*
                    centering ticklabels
                     */

                    if (scale.bandwidth && scale.discreteWidth) {
                        let _position = scale.discreteWidth(i) / 2 || 0,
                            position2d =
                                _.orient == 'bottom' || _.orient == 'top' ? `${_position + t[0]},${t[1]}` : `${t[0]},${_position + t[1]}`;
                        return `translate(${position2d})`;
                    }
                }
                return `rotate(${positionH.rotationInfo.rotate[0]} ${positionH.rotationInfo.rotate[1] + t[0]} ${positionH.rotationInfo
                    .rotate[2] + t[1]})`;
            },
            textAnchor: positionH.rotationInfo.textAnchor,
            // axis ticklabel custom transform
            customTransform: function(d, attr, val) {
                let scale = _.axisInfo.scale || tickInfo.scale,
                    callback = _.config.tickTransform;
                if (callback) {
                    let customTickTransform = invokeFunction(callback, d, scale, _.chart);
                    val += attr == 'x' ? customTickTransform[0] : customTickTransform[1];
                }
                return val;
            }
        },
        keys = {
            attrs_key: positionH.attrs_key
        };

    return { ...attrs, ...keys };
}

AxisRegistry.add('cartesian.ticklabelPosition', getTicklabelPosition);
