// @zohocharts/zc/src/components/race/raceComponents.js
import { getTspanTemplate } from '../../label/tspan-utils';
import { d3_select } from './../../core/d3-constants';
import { invokeFunction, isArray } from './../../utils/zcutils';

export function defaultRaceCompParser(data, raceSettings, chart) {
    var race = this;
    var _this = this;
    var raceSettings = race.userRaceSettings.components,
        labelConf = raceSettings.label,
        scrollLabelConf = labelConf.scroll,
        plotLabelConf = labelConf.plot,
        labelSettings = labelPropAndChidlren(scrollLabelConf, plotLabelConf, data, chart),
        buttonConf = raceSettings.button,
        buttonLeft = buttonConf.x,
        buttonTop = buttonConf.y;
    let components = [
        {
            parser: {
                mark: 'g',
                selector: '.plotLabel',
                data: plotLabelConf.show ? ['plot'] : [],
                order: '.seriesClippathholder',
                parent: chart.svg.selectAll('.chartgroup'),
                ...labelPropAndChidlren(scrollLabelConf, plotLabelConf, data, chart, "plot")
            },
            extender: plotLabelConf.markExtender
        },
        {
            parser: {
                mark: 'g',
                selector: '.scrollLabel',
                data: scrollLabelConf.show && chart.scroll ? ['scroll'] : [],
                order: '.custom-handle-path',
                parent: (chart.scroll && chart.scroll.brushInstance.customHandles) || chart.container,
                ...labelPropAndChidlren(scrollLabelConf, plotLabelConf, data, chart, "scroll")
            },
            extender: scrollLabelConf.markExtender
        },
        {
            parser: {
                mark: 'div',
            selector: '.zcbutton',
            data: buttonConf.show ? [1] : [],
            parent: d3.select(chart.svg.node().parentElement),
            config: {
                width: buttonConf.buttonSize + 'px',
                height: buttonConf.buttonSize + 'px'
            },
            properties: {
                attrs: {},
                styles: {
                    display: 'flex',
                    'place-content': 'center center',
                    'flex-wrap': 'wrap',
                    background: buttonConf.buttonColor,
                    'border-radius': '50px',
                    position: 'absolute',
                    margin: '-25px 0 0 -25px',
                    cursor: 'pointer',
                    top: buttonTop,
                    left: buttonLeft
                }
            },
            children: [
                {
                    mark: 'text',
                    selector: '.race_label',
                    data: buttonConf.label.show ? [data] : [],
                    config: {
                        html: function() {
                            //if want to give diff text, customise it here
                            return race.autoPlayMode;
                        }
                    },
                    properties: {
                        styles: {
                            'font-family': 'Source Sans Pro',
                            'font-size': '20px',
                            color: 'white',
                            'font-weight': 'bold',
                            'text-align': 'center',
                            padding: '0px 2px',
                            'line-height': 1,
                            ...buttonConf.label.styles
                        }
                    }
                },
                {
                    mark: 'div',
                    selector: '.raceButton-play',
                    data: ['play'],
                    properties: {
                        styles: {
                            overflow: 'hidden',
                            position: 'relative',
                            'border-radius': '20%',
                            transform: 'translateY(27.5%) translateX(7.5%) rotate(0deg) scaleX(.866)  skewY(30deg) ',
                            display: function() {
                                return race && race.autoPlayMode === 'pause' ? 'none' : null;
                            },
                            width: buttonConf.iconSize + 'px',
                            height: buttonConf.iconSize + 'px',
                            '--check-secondary': buttonConf.iconColor
                        }
                    }
                },
                {
                    mark: 'div',
                    selector: '.raceButton-pause',
                    data: ['pause'],
                    properties: {
                        styles: {
                            width: '0px',
                            height: buttonConf.iconSize + 'px',
                            'border-color': 'transparent transparent transparent ' + buttonConf.iconColor,
                            'box-sizing': 'border-box',
                            'border-style': 'double',
                            'border-width': '0px 0 0px ' + buttonConf.iconSize + 'px',
                            transform: 'translateY(0%) translateX(0%)  rotate(0deg) scaleX(.8) scaleY(.8)',
                            cursor: 'pointer',
                            display: function() {
                                return race && race.autoPlayMode === 'pause' ? null : 'none';
                            },
                            '--check-secondary': 'red'
                        }
                    }
                }
            ],
            onEnd: function(d) {
                var buttomLabelEle = d.selectAll('.race_label').node(),
                    conf = buttomLabelEle && (buttomLabelEle.getBBox ? buttomLabelEle.getBBox() : buttomLabelEle.getBoundingClientRect()),
                    buttonSize = buttonConf.buttonSize;
                if(buttomLabelEle){
                    d.style('width', conf.width + buttonSize + 'px').style('height', buttonSize + 'px');
                }

                d.on('click touchstart', async function() {
                    try{
                        //without below cond - even before chart init animation ending, when play button clciked, annotation and line not in sink.
                        if((chart.chartRendered || (chart.racing.started && !chart.racing.interrupt) || chart.racing.ended)/* && !chart.racing.interrupt*/){
                            var _this = chart.racing;
                            chart.racing.started = true;
                            chart.racing.raceMode = 'auto';
                            if (!_this.visibleFrameStatus) {
                                chart.racing.visibleFrameStatus = {
                                    isTransitionStarted: false,
                                    isRenderingStarted: true
                                };
                            }
                            var playMode = _this.autoPlayMode,
                                isPlay = playMode === 'play',
                                isReplay = playMode === 'replay',
                                isPause = playMode === 'pause';
                            if (isPlay || isReplay) {
                                chart.racing.preventEvent = true;
                                chart.racing.ended = false;
                                if (isReplay) {
                                    _this.replay = true;
                                }
                                _this.autoPlayMode = 'pause';
                                invokeFunction(_this.events.onInterrupt, _this, chart);
                                await _this.initiateRacing.call(_this, chart);
                                if (chart.racing) {
                                    if (chart.racing.allowRedraw) {
                                        var frameInfo = _this.keyFrameInfo,
                                            frameData = _this.data,
                                            _frame = frameData[frameInfo.index];
                                        chart.racing.allowRedraw = false;
                                        _this.autoPlayMode = 'play';
                                        invokeFunction(_this.events.onInterrupt, _this, chart);
                                    }
                                    chart.racing.started = false;
                                    chart.racing.ended = true;
                                    chart.racing.preventEvent = false;
                                }
                            } else if (isPause) {
                                chart.racing.interrupt = true;
                                // chart.racing.scrollView = null;
                                _this.replay = false;
                                _this.autoPlayMode = 'play';
                                invokeFunction(_this.events.onInterrupt, _this, chart);
                            }
                        }
                    }catch{
                        
                    }
                });
            }
            },
            extender: buttonConf.markExtender
        }
    ];
    // config.updator = function(){
    //     console.log("updator");
    //     var _this = this,
    //         playMode = _this.autoPlayMode;
    //     // if (showLabel) {
    //         // _this.custom_ele.buttonGroup.selectAll('.race_label').node().innerText = conf.text[_this.autoPlayMode];
    //     // }
    //     var _b_grp = _this.custom_ele.buttonGroup;
    //     if (playMode === 'pause') {
    //         _b_grp.selectAll('.raceButton-play').style('display', 'none');
    //         _b_grp.selectAll('.raceButton-pause').style('display', null);
    //     } else {
    //         _b_grp.selectAll('.raceButton-play').style('display', null);
    //         _b_grp.selectAll('.raceButton-pause').style('display', 'none');
    //     }
    // }
    //  if(type === "plot"){
    //     config[0].properties.attrs.transform = () => {
    //         let x, y, rotate;
    //         x = chart.chartarea.width/2;
    //         y = chart.chartarea.height/2;;
    //         rotate = 'rotate(0)';

    //         return 'translate(' + x + ',' + y + ')' + rotate;
    //     }
    //  }else{
    //     config[0].properties.attrs.transform = () => {
    //         let y =  -50;
    //         return 'translate(0,' + y + ') rotate(0)';
    //     }
    //  }
    return components;
}

export function raceComponentRenderer(data, chart, update, componentsParserArr) {
    var race = this,
        renderDom = $ZC.getComponent('mark.render');
    var raceCompParser = componentsParserArr,
        eventHandler = chart.eventHandler, oldBrush, activeBrush;

    if(eventHandler){
        oldBrush = eventHandler.oldBrushComponent,
        activeBrush = eventHandler.activeBrushComponent;
    if (!(oldBrush === null && activeBrush === null)) {
        eventHandler.oldBrushComponent = activeBrush;
        eventHandler.activeBrushComponent = 'scroll';
    }
      
    }
    raceCompParser.map(d => {
        d.parser = invokeFunction(d.extender, d.parser, race) || d.parser;
        renderDom(d.parser);
    });

    if(eventHandler){
        eventHandler.oldBrushComponent = oldBrush;
        eventHandler.activeBrushComponent = activeBrush;
    }
    
}

function labelPropAndChidlren(scrollLabelConf, plotLabelConf, data, chart, type) {
    var scrollTickBackGConf = scrollLabelConf.boundBox,
        plotTickBackGConf = plotLabelConf.boundBox;
    if(type === "plot"){
        return {
            properties: {
                attrs: {
                    transform: function(d, i) {
                        let x = invokeFunction(plotLabelConf.x, d, i),
                            y = invokeFunction(plotLabelConf.y, d, i);
                        return 'translate(' + x + ',' + y + ')';
                    }
                }
            },
            children: [
                {
                    //rect-background
                    mark: 'rect',
                    selector: '.tickerRect',
                    data: [data],
                    config: {
                        strokeWidth: plotTickBackGConf['strokeWidth']
                    },
                    properties: {
                        styles: {
                            "cursor": "pointer",
                            fill: plotTickBackGConf['fill'],
                            stroke: plotTickBackGConf['stroke'],
                            opacity: plotTickBackGConf.show ? 1 : 0
                        }
                    }
                },
                {
                    //label
                    mark: 'text',
                    selector: '.tickerText',
                    data: [data],
                    config: {
                        html: function(d) {
                            var text = plotLabelConf.text;
                            if (!$ZC.getComponent('utils.isArray')(d)) {
                                return getTspanTemplate(text([d]));
                            }
                            return getTspanTemplate(text(d));
                        },
                        x: 0,
                        y: 0
                    },
                    properties: {
                        styles: {
                            "cursor": "pointer",
                            ...plotLabelConf.styles
                        }
                    },
                    onEnd: function(d, i) {
                        if (d.node() && d.node().previousSibling) {
                            var plotLabelPadding = plotLabelConf.boundBox.padding,
                                conf = d.node().getBBox ? d.node().getBBox() : d.node().getBoundingClientRect(),
                                padding = plotLabelPadding;
                            d.nodes().map(function(d, i) {
                                conf =  d.getBBox ? d.getBBox() : d.getBoundingClientRect();
                                d3.select(d.previousSibling)
                                    .attr('x', conf.x - padding)
                                    .attr('y', conf.y - padding)
                                    .attr('width', conf.width + 2 * padding)
                                    .attr('height', conf.height + 2 * padding);
                            });
                        }
                    }
                }
            ]
        };
    }else{
        return {
            properties: {
                attrs: {
                    transform: function(d, i) {
                        let x = invokeFunction(scrollLabelConf.x, d, i),
                            y = invokeFunction(scrollLabelConf.y, d, i);
                        return 'translate(' + x + ',' + y + ')';
                    }
                }
            },
            children: [
                {
                    //rect-background
                    mark: 'rect',
                    selector: '.tickerRect',
                    data: [data],
                    config: {
                        strokeWidth: scrollTickBackGConf['strokeWidth']
                    },
                    properties: {
                        styles: {
                            "cursor": "pointer",
                            fill: scrollTickBackGConf['fill'],
                            stroke: scrollTickBackGConf['stroke'],
                            opacity: scrollTickBackGConf.show ? 1 : 0
                        }
                    }
                },
                {
                    //label
                    mark: 'text',
                    selector: '.tickerText',
                    data: [data],
                    config: {
                        html: function(d) {
                            var text = scrollLabelConf.text;
                            if (!$ZC.getComponent('utils.isArray')(d)) {
                                return getTspanTemplate(text([d]));
                            }
                            return getTspanTemplate(text(d));
                        },
                        x: 0,
                        y: 0
                    },
                    properties: {
                        styles: {
                            "cursor": "pointer",
                            ...scrollLabelConf.styles
                        }
                    },
                    onEnd: function(d, i) {
                        if (d.node() && d.node().previousSibling) {
                            var scrollLabelPadding = scrollLabelConf.boundBox.padding,
                                conf =  d.node().getBBox ? d.node().getBBox() : d.node().getBoundingClientRect(),
                                padding = scrollLabelPadding;
                            d.nodes().map(function(d, i) {
                                conf = d.getBBox ? d.getBBox() : d.getBoundingClientRect();
                                d3.select(d.previousSibling)
                                    .attr('x', conf.x - padding)
                                    .attr('y', conf.y - padding)
                                    .attr('width', conf.width + 2 * padding)
                                    .attr('height', conf.height + 2 * padding);
                            });
                        }
                    }
                }
            ]
        };
    }
    
}

// function getTspanTemplate(showAs) {
//     var newShowAs = '';
//     if (isArray(showAs)) {
//         showAs.forEach(function(d, i) {
//             newShowAs += `<tspan class='zcTspan'${i ? " dy='1.1em'" : ''}>${updateTspanTemplate(d)}</tspan>`;
//         });
//     } else {
//         var tagIndex = showAs.indexOf('<tspan');
//         if (tagIndex != -1) {
//             newShowAs = `<tspan class='zcTspan'>${updateTspanTemplate(showAs)}</tspan>`;
//         } else {
//             newShowAs = showAs;
//         }
//     }
//     return newShowAs;
// }

// function updateTspanTemplate(showAs) {
//     var str = '',
//         lastIndex = 0,
//         endTagLength = 8,
//         t = {
//             '<tspan': [],
//             '</tspan>': []
//         };

//     showAs.replace(/<tspan|<\/tspan>/g, function() {
//         t[arguments[0]].push(arguments[1]);
//     });

//     t['<tspan'].forEach(function(d, i) {
//         if (lastIndex != d) {
//             str += `<tspan>${showAs.substring(lastIndex, d)}</tspan>`;
//         }
//         str += showAs.substring(d, (lastIndex = t['</tspan>'][i] + endTagLength));
//     });

//     if (lastIndex != showAs.length) {
//         str += `<tspan>${showAs.substring(lastIndex, showAs.length)}</tspan>`;
//     }

//     return str;
// }
