// @zohocharts/zc/src/components/race/race-config.js
import { CHART_DOMAIN_HELPERS } from '../../brush/chart-helpers/domain-helpers';
import { simpleClone } from './../../utils/zcutils';

export function raceSettings(chart) {
    return {
        duration: 250,
        events: {
            onFrameInit: async function (race, chart) {
                var comps = [];
                comps.push(race.componentsParserArray[0]);
                comps.push(race.componentsParserArray[1]);
                race.compRenderer.call(race, chart.seriesdata, chart, null, comps);
            },
            onFrameLoad: (race, chart) => {
                var comps = [];
                comps.push(race.componentsParserArray[0]);
                comps.push(race.componentsParserArray[1]);
                race.compRenderer.call(race, chart.seriesdata, chart, null, comps);
            },
            onFrameInterrupt: (race, chart) => {
                //gets called on redraw, on pause and play of race, on click of scroll bar when race is playing.
                race.compRenderer.call(race, chart.seriesdata, chart, null, race.componentsParserArray);
            }
        },
        frames: {
            metadata: {
                column: {
                    datatype: 'time',
                    time: {}
                },
                racingColumns: [1] //when a column with string is given for all inter frames, to frame string will be got, if not given here,m for all iter frames from frame string will bw got by default
            },
            intermediateFrames: {
                count: 0,
                customInterpolator: null,
                dataSort: function(d, i, args) {
                    var sortType = 'descending',
                        sortByIndex = 1;
                    d[0].sort(function(a, b) {
                        return d3[sortType](a[sortByIndex], b[sortByIndex]);
                    });
                    //from and to frame has to be sorted by the user.while givng the frames itself.
                    return d;
                },
                _constructor: function(d, i, args) {
                    var frame = simpleClone(args.fromFrame);
                    frame.data = d;
                    return frame;
                }
            }
        },
        components: {
            label: {
                plot: {
                    show: true,
                    x: () => chart.chartarea.width / 2,
                    y: () => chart.chartarea.height / 2,
                    text: d => {
                        return ['<tspan>' + d[0].keyFrame_Id + '</tspan>'];
                    },
                    styles: {
                        'font-size': '15px', //check for callback
                        'font-family': 'Arial, Helvetica, sans-serif',
                        'fill': '#eee'
                    },
                    boundBox: {
                        show: false,
                        padding: 10,
                        fill: 'grey',
                        stroke: 'black',
                        strokeWidth: 1.5
                    }
                },
                scroll: {
                    show: true,
                    x: 0,
                    y: -50,
                    text: d => {
                        return ['<tspan>' + d[0].keyFrame_Id + '</tspan>'];
                    },
                    styles: {
                        'font-size': '20px',
                        'font-family': 'Arial',
                        'fill': 'red'
                    },
                    boundBox: {
                        show: false,
                        padding: 10,
                        fill: 'grey',
                        stroke: 'black',
                        strokeWidth: 1.5
                    }
                }
            },
            button: {
                show: true,
                buttonColor: 'grey',
                buttonSize: 35,
                iconColor: 'white',
                iconSize: 0.6 * 35,
                x: () => {
                    if (!chart.destroying) {
                        return 0.75 * chart.plotarea.left + 'px';
                    } else {
                        return 0;
                    }
                },
                y: () => {
                    if (!chart.destroying) {
                        return 0.65 * chart.plotarea.bottom + chart.plotarea.height + chart.plotarea.top + 'px';
                    } else {
                        return 0;
                    }
                },
                label: {
                    show: true,
                    styles: {
                        color: "white",
                        'font-family': 'Source Sans Pro',
                        'font-size': '20px',
                    }
                }
            },
            markExtender: null
        }
    };
}
