// @zohocharts/zc/src/bar/parser/radial/radialbar-parser.js
import { ParserRegistry } from "../../../core/parser-registry";
import { noOperation } from "../../../core/zc-constants";
import { renderDom } from "../../../mark/dom-render";
import { executeMarkOperators } from "../../../plot/mark-operator-executor";
import { createElement } from "../../../utils/zcsvgutils";
import { pick } from "../../../utils/zcutils";
import { ZC_BAR_MARKS_MAPPING } from "../../bar-marks-mapping";
import { radialbar_parser_helpers } from "./radialbar-parser-helpers";

export class RadialBarParser {
    //TODO: Work on enchancer later. Do radial bar rendering first. OK 👍
    #container; #mapping; #operators; #marks; #enhancer;
    constructor(args) {
        let { data, config, renderer, rendererIndex, legendOrder, paneid, chart, builderC, barDatasetIndex, phase, type, container, enhancer } = args;
        this.KEY = 'radialbar';
        this.data = data;
        this.config = config;
        this.paneid = paneid;
        this.rendererIndex = rendererIndex;
        this.barDatasetIndex = barDatasetIndex;
        this.conf = chart.rendererConf[rendererIndex];
        this.legendOrder = legendOrder;
        this.builderC = builderC;
        this.commonOption = {};
        this.chart = chart;
        this.type = type;
        this.phase = phase;
        this.renderer = renderer;
        this.markOperator = args.operator;
        // declare private variables here
        this.#container = container;
        this.#enhancer = enhancer;
        this.#initialize()
    }
    #initialize() {
        this.setMarkMapping();
        // this.commonOption = getBarCommonOption(this.conf, this.rendererIndex, this.legendOrder, this.paneid);
        // setRadialBarBuilderConfig.call(this);
        // setRadialBarRendererConfig.call(this);
        this.helpers = radialbar_parser_helpers.call(this);
        this.setMarkParser(['radialbar']);
    }

    #allowedMarkOperator(operator) {
        return operator ? (this.#operators = operator, operator) : this.#operators;
    }

    setMarkMapping(mapping) {
        this.#mapping = (mapping || ZC_BAR_MARKS_MAPPING).call(this);
    }
    setMarkParser(operator) {
        let operators = this.#allowedMarkOperator(operator),
            markMapping = this.#mapping,
            customOperator = this.markOperator;

        this.#collectEncode();
        this.#marks = executeMarkOperators.call(this, this.data, markMapping, operators, customOperator);
    }
    #collectEncode() {
        this.operators().forEach(operator => {
            let config = (ParserRegistry.config(operator) || noOperation);
            config.call(this);

            let encode = (ParserRegistry.encodes(operator) || noOperation);
            encode.call(this);
        });
        // radial_encode_collectors.call(this);
    }

    container() {
        return this.#container;
    }
    marks() {
        return this.#marks;
    }
    mapping() {
        return this.#mapping;
    }
    operators() {
        return this.#operators;
    }

    render(parent) {
        parent = parent || this.container();
        let holderIndex = pick(this.builderC.dynamicSeriesPosition, this.builderC.datasetIndex),
            index = this.conf.renderingOrderIndex();
        this.placeHolder = createElement('g', `.barseries_${holderIndex}`, [{ data: this.data, index }], parent);

        this.#marks.map(d => {
            renderDom(d.mark, d.holder || this.placeHolder)
        })
    }
}
ParserRegistry.parser('RadialBar', RadialBarParser);
