// @zohocharts/zc/src/behaviours/plot-zoom/chart/plot-zoom-handler-mappings.js
import { ChartZoomDataFinder } from '../../data-selection/chart/chart-zoom-data-finder';
import { filterChart_C } from '../../visual-effects/chart/controllers/filter-chart';
import { UpdateViewportArea_C } from '../../visual-effects/chart/controllers/update-viewport-area';
import { ClearOldInfo_C } from '../../visual-effects/chart/controllers/clear-old-info';
import { UPDATE_SCROLL_BY_DOMAIN } from '../../visual-effects/chart/controllers/scroll-domain';
import { Registry } from '../../../core/zc-registry';
import { d3_set } from '../../../core/d3-constants';
import { ChartZoomListener } from '../../listeners/chart-zoom-listener';
import { updateChartZoomTransform } from '../../visual-effects/chart/controllers/update-chart-zoom-transform';
import { $Browser } from '../../../core/browser-detect';
import { ShowBrushResetButton } from '../../visual-effects/chart/controllers/plot-reset-button';

let zoom_start_listeners = [];
if (!$Browser.isTouchDevice) {
    zoom_start_listeners = [ClearOldInfo_C];
}

const _chartZoomHandlers = {
    LISTENER: {
        start: ChartZoomListener,
        zoom: ChartZoomListener,
        end: ChartZoomListener
    },
    DATA_FINDER: {
        start: ChartZoomDataFinder,
        zoom: ChartZoomDataFinder,
        end: ChartZoomDataFinder
    },
    BEHAVIOUR: {
        start: zoom_start_listeners,
        zoom: [ClearOldInfo_C, filterChart_C, UPDATE_SCROLL_BY_DOMAIN],
        end: [UpdateViewportArea_C, ShowBrushResetButton]
    },
    LISTENER_TYPES: d3_set(['start', 'zoom', 'end']),
    SUPPORTED: chart => chart.isAxisCategory && !chart.isPolarAxisCategory,
    UPDATE: updateChartZoomTransform
};

Registry.setComponent('_chartZoomHandlers', _chartZoomHandlers);

export { _chartZoomHandlers };
