// @zohocharts/zc/src/behaviours/plot-zoom/chart/sankey-zoom-handler-mappings.js
import { ClearOldInfo_C } from '../../visual-effects/chart/controllers/clear-old-info';
import { Registry } from '../../../core/zc-registry';
import { d3_set } from '../../../core/d3-constants';
import { ChartZoomListener } from '../../listeners/chart-zoom-listener';
import { updateChartZoomTransform } from '../../visual-effects/chart/controllers/update-chart-zoom-transform';
import { $Browser } from '../../../core/browser-detect';
import { applyZoomToSankeyPlot, applyZoomToSankeyLabels, clearOldHighlightSankey, clearCacheForEvents } from '../../visual-effects/chart/controllers/sankey-zoom-scale';

let zoom_start_listeners = [];
if (!$Browser.isTouchDevice) {
    zoom_start_listeners = [ClearOldInfo_C];
}

const _sankeyZoomHandlers = {
    LISTENER: {
        start: ChartZoomListener,
        zoom: ChartZoomListener,
        end: ChartZoomListener
    },
    DATA_FINDER: {},
    BEHAVIOUR: {
        start: [applyZoomToSankeyPlot, applyZoomToSankeyLabels, clearOldHighlightSankey],
        zoom: [applyZoomToSankeyPlot, applyZoomToSankeyLabels],
        end: [applyZoomToSankeyPlot, applyZoomToSankeyLabels, clearCacheForEvents]
    },
    LISTENER_TYPES: d3_set(['start', 'zoom', 'end']),
    SUPPORTED: chart => chart.dataObject.isFlowChartCategory && chart.plot.renderer.mode == 'svg',
    UPDATE: updateChartZoomTransform
};

Registry.setComponent('_sankeyZoomHandlers', _sankeyZoomHandlers);

export { _sankeyZoomHandlers };
