// @zohocharts/zc/src/legend/base-legend.js
//
import { ORDINAL_DATATYPE, TIME_DATATYPE } from '../core/datatype-constants';
import { LIST_LEGEND_TYPE } from './legend-type-constants';
import { BaseView } from '../view/base-view';
import { filterCategories } from '../axis/tick-labels';
import { d3_scaleOrdinalOrg } from '../core/d3-constants';
import { defined } from '../utils/zcutils';
import { hasAutoFormat } from '../axis/tick-helpers';
import { getAutoTimeSubFunction } from '../formatter/formatter';
import { getEntries, getTickCount, getUserTickCount } from './size-legend/size-legend-helpers';

class BaseLegend extends BaseView {
    resize(width, height) {
        let bound = this.spec.bound;
        bound.maxWidth = width;
        bound.maxHeight = height;

        this.create();
        this.draw(this.element);
    }

    viewScaleComponent() {
        let spec = this.spec,
            { config, scaleComponent, view } = spec,
            scale = scaleComponent.getScale(),
            range = scale.range(),
            domain = scale.domain(),
            size = range[range.length - 1],
            domainSchema = scaleComponent.schema.domain,
            { meta, dataType } = scaleComponent.schema.data,
            { datatype: originalDataType } = meta,
            formatConfig = meta[originalDataType] || {},
            _ranges = domainSchema.ranges,
            _stops = domainSchema.stops,
            userTickCount = getUserTickCount(config, scale, view),
            tickCount = getTickCount(config, scale, view),
            defaultReverse = view === LIST_LEGEND_TYPE,
            ticks,
            labels,
            viewScale;

        // Handling subfunction = 'AUTO' && format.specifier = 'auto'
        if (dataType === TIME_DATATYPE) {
            updateTimeMeta(formatConfig, domain);
        }

        if (dataType === ORDINAL_DATATYPE) {
            if (!config.ranges) {
                config.ranges = domain.map(d => [d]);
            }
        }

        if (_ranges || dataType === ORDINAL_DATATYPE) {
            ticks = dataType === ORDINAL_DATATYPE ? domain : _ranges || domain; // For ordinal, the ticks should be 1D array
            viewScale = d3_scaleOrdinalOrg()
                .domain(ticks)
                .range(range);
        } else {
            ticks = _stops ? domain : getEntries(scale, dataType, tickCount, formatConfig.subfunction, defaultReverse);
            viewScale = scale;
        }

        if (userTickCount) {
            labels = filterCategories(ticks, userTickCount);
        }

        return {
            ticks,
            labels: labels || ticks,
            scale: viewScale,
            bound: { width: size, height: size },
            labelData: _ranges && ticks._dummy && ticks.length === 1 && ticks[0][0] === ticks[0][1] ? [[ticks[0][0]]] : null // ZC2548
        };
    }
}

function updateTimeMeta(formatConfig, domain) {
    let subfunction = formatConfig.subfunction,
        autoFormat = hasAutoFormat(formatConfig, 'legend');

    if (!defined(subfunction) || subfunction === 'AUTO' || autoFormat) {
        var info = getAutoTimeSubFunction(domain[1] - domain[0]);
        formatConfig.subfunction = info.subFunction;
    }

    if (autoFormat) {
        (formatConfig.legendformat || formatConfig.format).specifier = null;
    }
}

export { BaseLegend };
