// @zohocharts/zc/src/legend/legend-utils.js
//$Id$
//legend-utils.js

import { mathMin } from '../core/zc-constants';
import { generateDefaultPatternTheme } from '../effects/patternThemes';
import { d3_map, d3_range } from '../core/d3-constants';
import { getMonochromeColor, invokeFunction, bbox, defined } from '../utils/zcutils';
import { getLegendBound } from './legend-bound';
import { _imageFilePath } from '../core/zcconf';
import { isSvgElement } from '../utils/zcsvgutils';
import { LEFT_ALIGN, TOP_ALIGN } from '../core/align-constants';
import { ZC_LEGEND_maxCenterHeight, ZC_LEGEND_maxCenterWidth, ZC_LEGEND_maxHeightHoriRatio, ZC_LEGEND_maxHeightVertRatio, ZC_LEGEND_maxWidthHoriRatio, ZC_LEGEND_maxWidthVertRatio } from './legend-constants';

export var legutil = function() {};
legutil.colorpallete = d3_map();

legutil.colorpallete.set('greenish', [
    '#6c6c6d',
    '#1ba98e',
    '#00ff80',
    '#00ffc0',
    '#00ffff',
    '#40ff00',
    '#40ff40',
    '#40ff80',
    '#40ffc0',
    '#40ffff',
    '#80ff00',
    '#80ff40',
    '#80ff80',
    '#80ffc0',
    '#80ffff'
]);
legutil.colorpallete.set('pinkish', [
    '#ff00ff',
    '#ff00c0',
    '#ff0080',
    '#ff0040',
    '#ff0000',
    '#c000ff',
    '#c000c0',
    '#c00080',
    '#c00040',
    '#c00000',
    '#8000ff',
    '#8000c0',
    '#800080',
    '#800040',
    '#800000'
]);
legutil.colorpallete.set('countries', ['#00c6e4', '#faa43a', '#5da5da', '#60bd68', '#f17cb0', '#b2912f', '#307d99']);
legutil.colorpallete.set('saturate5', ['#594f4f', '#547980', '#45ada8', '#97843a', '#518058', '#975156']);
legutil.colorpallete.set('ocean5', ['#ff7f00', '#6a4a3c', '#cc333f', '#eb6841', '#edc951', '#975156']);
legutil.colorpallete.set('zoho-reports-old', [
    '#FF465B',
    '#04C36B',
    '#3A92E9',
    '#FDC33A',
    '#D65587',
    '#FC7C63',
    '#51A5A9',
    '#BFBFBF',
    '#91CD6A',
    '#BE9B89',
    '#7D8DB0',
    '#B68E2E',
    '#C6D5B0',
    '#D1ACCE',
    '#E0AE60',
    '#5CD29F',
    '#716398',
    '#AD8686',
    '#FF99A8',
    '#A0C0C0',
    '#2D7094',
    '#553D4E',
    '#B2449D',
    '#FED577',
    '#D3D159',
    '#AAAAAA',
    '#7ED9D9',
    '#F7A26D',
    '#8ABFA2',
    '#767676',
    '#E36068',
    '#A3AFC7'
]);
legutil.colorpallete.set('zoho-reports-new', [
    '#60bfdf',
    '#ffa060',
    '#aadf80',
    '#ff6c7d',
    '#c9ac66',
    '#bf80bf',
    '#df6060',
    '#606060',
    '#60a0df',
    '#ff99a8',
    '#feff60',
    '#a0dfdf',
    '#dfa080',
    '#dfdfdf',
    '#a0a060',
    '#a080bf',
    '#a06060',
    '#5cd29f',
    '#ffbe60',
    '#bfa080',
    '#f28fb5',
    '#8080bf',
    '#bfbfbf',
    '#dfdf80',
    '#80a060',
    '#80a060',
    '#a08060',
    '#7ed9d9',
    '#be9b89',
    '#bf6060',
    '#a0bf60',
    '#fc7c63'
]);
legutil.colorpallete.set('colors', ['#0eaadd', '#1ac9bb', '#92c658', '#dbbe56', '#ee7d54', '#ef4f6a', '#b952ef']);
legutil.colorpallete.set('flat-ui-colors', [
    '#23a96e',
    '#bca634',
    '#f75a51',
    '#39b6c5',
    '#f9ce8d',
    '#983a67',
    '#634c3f',
    '#d7dab9',
    '#357180',
    '#983a67',
    '#b16b76',
    '#7f90a0'
]);
legutil.colorpallete.set('MEMobile', [
    '#73c2f4',
    '#ecd543',
    '#e6be3f',
    '#da7b36',
    '#c84b37',
    '#9c240f',
    '#9bcedd',
    '#ffe560',
    '#ffe560',
    '#c67c99',
    '#c6db4c'
]);
legutil.colorpallete.set('Toddler', [
    '#f8d0db',
    '#9be1ed',
    '#c4ead1',
    '#fbe2aa',
    '#e5e5e5',
    '#d6b65f',
    '#be9fbf',
    '#ffe560',
    '#ffe560',
    '#c67c99',
    '#c6db4c'
]);
legutil.colorpallete.set('BrightNDull', [
    '#47bac1',
    '#8ccace',
    '#72c3f5',
    '#71bfee',
    '#ac66ad',
    '#c5a0c6',
    '#906aa6',
    '#b29dbf',
    '#fd9f94',
    '#f6cac4',
    '#ff759c',
    '#f6acc1'
]);
legutil.colorpallete.set('BlueShades', ['#00c6e4', '#6ad4e4', '#a3dbe4', '#c3e0e4', '#e4e4e4', '#c8c8c8', '#aac4c8']);
legutil.colorpallete.set('graphite6', ['#595959', '#767676', '#373737', '#9d9d9d', '#bfbfbf', '#222222']);
legutil.colorpallete.set('chalk', ['#FFFFFF', '#FAFAFA', '#CCCCCC']);
legutil.colorpallete.set('map', ['#FB0D0D', '#53B808', '#B80894', '#2ADBDD']);

legutil.patternPallete = d3_map();
legutil.imagePallete = d3_map();

legutil.imagePallete.set('linesndots', [
    '01.png',
    '02.png',
    '03.png',
    '04.png',
    '05.png',
    '06.png',
    '07.png',
    '08.png',
    '09.png',
    '10.png',
    '11.png',
    '12.png',
    '13.png',
    '14.png',
    '15.png',
    '16.png',
    '17.png',
    '18.png',
    '19.png',
    '20.png'
]);

legutil.imagePallete.set('dotsfirst', ['13.png', '12.png', '13.png', '14.png', '15.png', '16.png', '17.png', '18.png', '19.png', '20.png']);

legutil.imagePallete.set('stones', ['01.jpg', '02.jpg', '03.jpg']);

generateDefaultPatternTheme();
legutil.patternPallete.set('stripes-fill', [
    'hatchfill',
    'hatchDashfill',
    'checkerBoard',
    'horizontalStripesfill',
    'verticalStripesfill',
    'crossStripesfill'
]);
legutil.patternPallete.set('stripes', [
    'hatch',
    'thinThickHatch',
    'hatchLeft',
    'hatchDash',
    'horizontalStripes',
    'crossStripes',
    'verticalStripes',
    'thickHatch',
    'thinHatch',
    'thinVerticalStripes',
    'thinHorizontalStripes'
]);
legutil.patternPallete.set('dotted', ['circle', 'tinyCircle', 'crossCircle', 'floatingCircle', 'dot']);
legutil.patternPallete.set('dotted-fill', ['circlefill', 'crossCirclefill', 'floatingCirclefill', 'dotfill']);
legutil.patternPallete.set('shapes', ['thinThickHatch', 'crossCircle', 'star', 'wave', 'cap', 'nylon']);
legutil.patternPallete.set('shapes-fill', [
    'checkerBoard',
    'floatingCirclefill',
    'squarefill',
    'starfill',
    'wavefill',
    'capfill',
    'nylonfill'
]);

export var setLegendColorPallete = function(name, value) {
    legutil.colorpallete.set(name, value);
};

export var getLegendColorPallete = function(name) {
    return legutil.colorpallete.get(name);
};

export var setLegendPatternPallete = function(name, value) {
    legutil.patternPallete.set(name, value);
};

export var getLegendPatternPallete = function(name) {
    return legutil.patternPallete.get(name);
};

export var setLegendImagePallete = function(name, value) {
    legutil.imagePallete.set(name, value);
};

export var getLegendImagePallete = function(name) {
    return legutil.imagePallete.get(name);
};

// $ZC.legutil.patternPallete.set("stripes", ["Default00001.png", "Default00002.png", "Default00003.png", "Default00004.png", "Default00005.png", "Default00006.png", "Default00007.png", "Default00008.png",
//     "Default00009.png", "Default00010.png", "Default00011.png", "Default00012.png", "Default00013.png", "Default00014.png", "Default00015.png", "Default00016.png",
//     "Default00017.png", "Default00018.png", "Default00019.png", "Default00020.png", "Default00021.png", "Default00022.png"
// ]);
//append relative pattern filepath
appendFilePath();

legutil.getMonochromeColorPallette = function(conf) {
    var domain = [0, +conf.count - 1],
        range = [+conf.colorGamma[0], +conf.colorGamma[1]];
    var monoClrGen = getMonochromeColor(conf.baseColor, domain, range);
    return d3_range(conf.count).map(function(d, i) {
        return monoClrGen(i);
    });
};

function getColorPallete(palletename, value) {
    if (legutil.colorpallete.has(palletename)) {
        return legutil.colorpallete.get(palletename);
    } else {
        legutil.colorpallete.set(palletename, value);
    }
}

function getPatternPallete(palletename, value) {
    if (legutil.patternPallete.has(palletename)) {
        return legutil.patternPallete.get(palletename);
    } else {
        legutil.patternPallete.set(palletename, value);
    }
}

function getImagePallete(imagename, value) {
    if (legutil.imagePallete.has(imagename)) {
        return legutil.imagePallete.get(imagename);
    } else {
        legutil.imagePallete.set(imagename, value);
    }
}

function appendFilePath() {
    legutil.imagePallete.each(function(value, pallete) {
        var arr = value.map(function(d, i) {
            return _imageFilePath + d;
        });
        legutil.imagePallete.set(pallete, arr);
    });
}

/**
 *
 * @param {{width:number, height:number, left: number, top: number}} area
 * @param chartObj
 * @param considerMargin
 * @return {{x: number, y: number}}
 */
export function getLegendPosition(area, chartObj, considerMargin, element, legendConfig) {
    let isSingleLayout = chartObj.legend.isSingleLayout; //Remove
    element = element || chartObj.legendarea.element;
    legendConfig = legendConfig || chartObj.systemConf.legend;

    var legNode = element.node(),
        alignment = getLegendAlignment(legendConfig),
        left = 0,
        top = 0,
        getMargin = function(side) {
            if (considerMargin === false) {
                return 0;
            }
            return +legendConfig[side];
        },
        isContinuousLegend = isSvgElement(legNode),
        legendBBox = isContinuousLegend ? bbox(legNode) : legNode.getBoundingClientRect(),
        legendWidth = legendBBox.width,
        legendHeight = legendBBox.height,
        floating = alignment.floating,
        canvasArea = chartObj.canvasarea,
        { maxWidth, maxHeight } = getLegendBound(legendConfig, chartObj.chartarea);

    if (!isContinuousLegend) {
        legendBBox.x = 0;
        legendBBox.y = 0;
    }

    if (floating) {
        left = legendConfig.x - legendBBox.x;
        top = legendConfig.y - legendBBox.y;

        left = mathMin(canvasArea.width - mathMin(legendBBox.width, maxWidth), left);
        top = mathMin(canvasArea.height - mathMin(legendBBox.height, maxHeight), top);

        return [left, top];
    }

    switch (alignment.hAlign) {
        case 'left': {
            left = area.left + getMargin('marginLeft') - legendBBox.x;
            break;
        }
        case 'center': {
            left = area.width / 2 - legendWidth / 2 - legendBBox.x;
            break;
        }
        case 'right': {
            // TODO #ZC1506
            var l = isContinuousLegend || !isSingleLayout ? area.left : 0;
            left = l + area.width - legendWidth - getMargin('marginRight') - legendBBox.x;
            break;
        }
    }

    switch (alignment.vAlign) {
        case 'top': {
            top = area.top + getMargin('marginTop') - legendBBox.y;
            break;
        }
        case 'center': {
            top = area.top + area.height / 2 - legendHeight / 2 - (!isSingleLayout ? legendBBox.y : 0);
            break;
        }
        case 'bottom': {
            top = area.top + area.height - legendHeight - getMargin('marginBottom') - legendBBox.y;
        }
    }

    return [left, top, legendWidth, legendHeight];
}

//! Note: Migrated from legend-helper to legend-utils for custom bundle
/**
 *
 * @param legendConfig
 * @returns {{hAlign: *, vAlign: *, layout: *}}
 */
export function getLegendAlignment(config) {
    // fix center top
    var { hAlign, vAlign, layout: view, x, y } = config,
        floating = defined(x) && defined(y);

    // keep existing support
    if (hAlign === 'center' && vAlign === 'center') {
        hAlign = 'right';
    }

    if (floating) {
        hAlign = LEFT_ALIGN;
        vAlign = TOP_ALIGN;
    }

    if (!defined(view)) {
        if (floating) {
            view = 'vertical';
        } else if (vAlign === 'top' || vAlign === 'bottom') {
            view = 'horizontal';
        } else {
            view = 'vertical';
        }
    }

    return {
        hAlign: hAlign,
        vAlign: vAlign,
        layout: view,
        floating: floating
    };
}

/**
 *
 * @param legendOptions
 * @return {{width: number, height: number}}
 */
export function getDefaultLegendMaxDimension(legendOptions) {
    var legendAlign = getLegendAlignment(legendOptions),
        vAlign = legendAlign.vAlign,
        hAlign = legendAlign.hAlign,
        isHorizontalLayout = legendAlign.layout === 'horizontal',
        widthRatio = 0,
        heightRatio = 0;

    if (isHorizontalLayout) {
        if (vAlign === 'top' || vAlign === 'bottom') {
            widthRatio = ZC_LEGEND_maxWidthHoriRatio;
            heightRatio = ZC_LEGEND_maxHeightHoriRatio;
        } else if (vAlign === 'center') {
            widthRatio = ZC_LEGEND_maxCenterWidth;
            heightRatio = ZC_LEGEND_maxHeightVertRatio;
        }
    } else {
        if (hAlign === 'left' || hAlign === 'right') {
            widthRatio = ZC_LEGEND_maxWidthVertRatio;
            heightRatio = ZC_LEGEND_maxHeightVertRatio;
        } else if (hAlign === 'center') {
            widthRatio = ZC_LEGEND_maxWidthHoriRatio;
            heightRatio = ZC_LEGEND_maxCenterHeight;
        }
    }

    return {
        width: widthRatio,
        height: heightRatio
    };
}

export function getLegendShowAsText(d, i, options, chartObj) {
    var showAs = options.showAs,
        seriesPos = 0,
        itemPos = i;

    if (chartObj.isAxisCategory || chartObj.isNonAxisMultiSeriesCategory || chartObj.isGeoMap) {
        seriesPos = i;
        itemPos = 0;
    }

    // added function support in showAs
    if (showAs) {
        showAs = invokeFunction(showAs, d, i, chartObj);
        return chartObj.dataset.getCustomizedVal(chartObj, showAs, seriesPos, 0, itemPos, d, 'legend');
    }
}
