// @zohocharts/zc/src/legend/legend-init.js
//$Id$
//legend-init.js
import { legend } from '../core/main-zc-init';
import { Registry } from '../core/zc-registry';
import { Layout } from '../layout/layout';
import { stateChangeListener } from '../events/eventActions';
import { collect, render } from './legend-renderer-helpers';
import { getLegendPosition } from './legend-utils';
import { updateChartAndPlotBound } from './legend-bound';
import { simpleClone } from '../utils/zcutils';
import { addLegendApi } from './legend-api';
import { SERIES_LEGEND_CLASSNAME } from './legend-constants';

/**
 * 
 * @param {any} chartObj 
 * @param {object} spec - custom specs provided by user
 * @param {string} spec.parent - parent container css selector of the legend element
 */
legend.render = function (chartObj, spec) {
    let legendConfig = chartObj.systemConf.legend;
    // call event on legend init
    stateChangeListener(chartObj, legendConfig.events, null, 'init');

    let legendMeta = collect(chartObj, spec);
    let legends = legendMeta.map(render).filter(_ => !!_);
    addLegendApi(legends, chartObj);

    let colorLegend = legends.get('clr-0');
    chartObj.legendarea = getLegendArea(chartObj, colorLegend); // Adapting old method // Remove

    // Save initial chartarea, plotarea to use on expandable click
    chartObj._chartarea = simpleClone(chartObj.chartarea);
    chartObj._plotarea = simpleClone(chartObj.plotarea);

    legend.align(legends, chartObj, chartObj.chartarea, chartObj.plotarea);
    addEventApi(colorLegend, chartObj, legends); // Remove

    chartObj.legends = legends;

    // call legend on load event
    // #ZC1898
    stateChangeListener(chartObj, legendConfig.events, null, chartObj.updating ? 'onreload' : 'onload');
};

legend.align = function (legends, chartObj, area, plotarea) {
    let legendConfig = chartObj.systemConf.legend;

    if (chartObj.legend.isSingleLayout) {
        positionSingleLayout(chartObj); // Remove
    } else {
        let options = {
            area,
            order: legendConfig.order,
            getPosition: (element, config, area) => getLegendPosition(area, chartObj, null, element, config),
            callbackArgs: chartObj
        };

        let layout = new Layout(legends, options);
        legends.layout = layout;
        layout.areaToBeUpdated.push(plotarea);
        layout.create();
    }
};

export function getLegendArea(chartObj, comp) {
    if (!comp) {
        return {};
    }

    let element = comp.element,
        legendArea = {
            element,
            show: comp.enabled
        };

    if (comp.enabled) {
        let xy = getLegendPosition(chartObj.chartarea, chartObj, null, element);
        legendArea.x = xy[0];
        legendArea.y = xy[1];
        legendArea.width = xy[2];
        legendArea.height = xy[3];
    }

    return legendArea;
}

function addEventApi(colorLegend, chartObj, legends) {
    if (colorLegend && !colorLegend.enabled && colorLegend.meta.view === 'list') {
        chartObj.legend.component = legend.discrete.get(colorLegend.meta);
        legend.events(chartObj, chartObj.legendarea.element.selectAll('li.series'), {
            allSelected: true,
            isFirstClickAfterDeselection: false //TODO: check
        });
    }
    else if (colorLegend && colorLegend.enabled && colorLegend.meta.view === 'band') {//to add API in the colorBand element
        chartObj.legend.component = legend.discrete.get(colorLegend.meta);
        legend.events(chartObj, chartObj.legendarea.element.selectAll('li.series'), {
            allSelected: true,
            isFirstClickAfterDeselection: false 
        });
    }
}

function positionSingleLayout(chartObj) {
    let { x, y, element, show } = chartObj.legendarea;
    if (show) {
        Layout.positionElement(element, x, y);
        updateChartAndPlotBound(chartObj);
    }
}

export let _legendRenderer = legend.render;
Registry.setComponent('legend.render', _legendRenderer);
