// @zohocharts/zc/src/legend/legend-continuous-events.js
//$Id$
//legend-continuous-events.js
// import { HighlightSeries } from './events/visual-effects/highlight-series';
// import { ResetHighlightSeries } from './events/visual-effects/reset-highlight-series';
import { legend } from '../core/main-zc-init';
import { getLegendAlignment } from './legend-utils';
import { appendEle } from '../utils/zcsvgutils';
import { NONE, NULL, POINTER, UNDEFINED } from '../core/zc-constants';
import { validateNullVal, defined, jsonStringify, pInt } from '../utils/zcutils';
import { cancelAnimation } from '../utils/animationutils';
import { getSvgTransform } from '../utils/svg-transforms';
import { getFormattedValue } from '../formatter/formatter';
import { attachevents, getMousePosition } from '../events/eventActions';
import { d3_event } from '../core/d3-constants';
import { getHighlightedItemFromMousePoint } from './events/data-selection/size-slider-data-finder';
import { getHoverOptions } from './events/legend-event-helpers';
import { _modularized } from '../core/zcconf';

export * from './legend-continuous-highlight-slider';
export * from './legend-continuous-highlight-tooltip';
export * from './legend-brush-options';

legend.continuous.events = function(chartObj, options, legendComponent) {
    chartObj.legend.highlight = {};

    var spec = options.spec,
        { config } = spec,
        container = options.element,
        legendOptions = config,
        hoverOptions = getHoverOptions(legendOptions, 'clr'),
        element = container.selectAll(legendOptions.filter.enabled ? '.selection' : '.colorband'),
        position = options.position || getSvgTransform(container.selectAll('.zc-band')).translate,
        highlightTrackerElement = container.selectAll('.zcbrushhighlighttracker'),
        legendAlignment = getLegendAlignment(legendOptions),
        isHorizontal = legendAlignment.layout === 'horizontal',
        highlightElement = appendEle(container, 'g', [1], '.', 'class', 'highlight').style('display', NULL), // Setting "display": null is important to Firefox when hoverOptions.type is undefined on draw and will get defined on redraw
        indicatorObj = (chartObj.legend.highlight.slider = legend.continuous.slider(chartObj, highlightElement, spec)),
        tooltipObj = (chartObj.legend.highlight.tooltip = legend.continuous.tooltip(chartObj, highlightElement, config)),
        sliderElement = indicatorObj.container,
        events_legend = legendOptions.events,
        eventDisabled,
        highlightedItem,
        previousHighlightedItem;
    // manager = {
    //     chart: chartObj,
    //     options: { config, highlightManager: chartObj.legendHighlight },
    //     previousHighlightedItem: NULL
    // };

    highlightElement.style('display', NONE).style('opacity', 0);

    if (events_legend) {
        eventDisabled = validateNullVal(events_legend.enabled, true) == false;
    }

    if (!_modularized.colorBandLegend && !eventDisabled && hoverOptions.type && hoverOptions.type !== NONE) {
        var eventAttacherOptions = {
                chartObj: chartObj
            },
            mouseHandlers = {
                mousemove: defaultOnMouseOverFunc,
                mouseout: defaultOnMouseOutFunc
            },
            touchHandlers = NULL;
        attachevents(element, mouseHandlers, touchHandlers, eventAttacherOptions, 'highlight');
        attachevents(highlightTrackerElement, mouseHandlers, touchHandlers, eventAttacherOptions, 'highlight');
    }

    function defaultOnMouseOverFunc() {
        var mouse = getMousePosition((legendOptions.filter.enabled ? highlightTrackerElement : element).node(), d3_event()),
            pt = isHorizontal ? mouse[0] : mouse[1];
        highlightedItem = getHighlightData(mouse);

        // Hide brush tooltip
        d3_event().stopImmediatePropagation();

        //#ZC2273
        if (chartObj.legend.brush.tooltipComponent) {
            chartObj.legend.brush.tooltipComponent.hide(true);
        }

        legendComponent.highlightItem(pt, true);
        legendComponent.highlightSeries(highlightedItem, true);
    }

    function defaultOnMouseOutFunc(ele, d, i) {
        legendComponent.highlightItem(NULL, false);
        legendComponent.highlightSeries(previousHighlightedItem, false);
    }

    function getHighlightData(mouse) {
        let scale = chartObj.axes.clr.scale,
            scale_domain = scale.domain(),
            brushScale = _brushScale(),
            //
            options = {
                scale,
                isHorizontal,
                config: legendOptions,
                ranges: legendOptions.colorBand.ranges,
                reversed: legendOptions.colorAxis && legendOptions.colorAxis.reversed,
                brushScale,
                singleDomainContinuous: scale_domain[0] === scale_domain[1],
                encode: 'clr'
            };

        return getHighlightedItemFromMousePoint(mouse, options);
    }

    legendComponent.highlightItem = function(point, onLegHover, forced, item) {
        element.style('cursor', onLegHover ? POINTER : NULL);

        if (onLegHover) {
            highlight(point);
            showTooltip(defined(item) ? item : highlightedItem);
        } else {
            reset(forced);
        }
    };

    legendComponent.highlightSeries = function(highlightedItem, onLegHover) {
        chartObj.highlightContext = NULL;
        var highlightObj = chartObj.legendHighlight;

        if (onLegHover) {
            if (defined(previousHighlightedItem) && jsonStringify(previousHighlightedItem) !== jsonStringify(highlightedItem)) {
                var temp = previousHighlightedItem;
                previousHighlightedItem = NULL;
                legendComponent.highlightSeries(temp, false);
            }

            chartObj.seriesdata.forEach(function(d, j) {
                highlightObj.highlightSelectedSeries(legendOptions, j, highlightedItem);
                highlightObj.highlightUnselectedSeries(legendOptions, j, highlightedItem);
            });
            previousHighlightedItem = highlightedItem;
        } else {
            chartObj.seriesdata.forEach(function(d, j) {
                highlightObj.reset(legendOptions, j, highlightedItem, chartObj.seriesdata.length - 1 === j);
            });
        }
        // if (onLegHover) {
        //     HighlightSeries.call(manager, { point: highlightedItem });
        // } else {
        //     ResetHighlightSeries.call(manager);
        // }
    };

    function highlight(point) {
        var userRanges = legendOptions.colorBand.ranges,
            reversed = legendOptions.colorAxis && legendOptions.colorAxis.reversed,
            brushScale = _brushScale(),
            clr;

        if (userRanges && userRanges.length) {
            var bandWidth = brushScale.bandwidth(),
                index = pInt(point / bandWidth);

            index = (isHorizontal && reversed) || (!isHorizontal && !reversed) ? userRanges.length - 1 - index : index;
            point = brushScale(userRanges[index]) + bandWidth / 2;
            clr = chartObj.seriesColor[index % chartObj.seriesColor.length];
        } else {
            clr = chartObj.axes.clr.scale(brushScale.invert(point));
        }

        var trans = isHorizontal ? [point, 0] : [0, point];
        trans[0] += position[0];
        trans[1] += position[1];

        cancelAnimation(highlightElement);
        container.selectAll('.highlight').style('fill', clr);
        highlightElement
            // .transition()
            // .ease(d3_ease_linear)
            .style('display', NULL)
            .style('opacity', 1)
            .attr('transform', `translate(${trans})`);
    }

    function reset(forced) {
        if (forced) {
            highlightElement.style('opacity', 0);
        } else {
            highlightElement.transition().style('opacity', 0);
        }
    }

    function showTooltip(item) {
        var userRanges = legendOptions.colorBand.ranges,
            columnInfo = chartObj.dataset.getColumn('clr'),
            datatype = columnInfo.datatype,
            formatConf = columnInfo[datatype],
            value = userRanges && userRanges.length ? userRanges[item] : item[0];

        tooltipObj.show(getFormattedValue(formatConf, value, datatype, chartObj, 'legend', UNDEFINED, '-'));
    }

    function _brushScale() {
        return chartObj.axes.clr.brushScale;
    }
};

export var _legendContinuousEvents = legend.continuous.events;
