// @zohocharts/zc/src/legend/legend-symbols.js
//$Id$
//legend-symbols.js
/*
 * TODO:
 * Heatmap, charts with color scale
 */
// import { legend_helpers } from './legend-helper';
import * as constants from '../core/chartname-constants';
import { getSVGsymbol } from '../utils/svg-symbols';
import { legend, legend_helpers } from '../core/main-zc-init';
import { roundedRect } from '../utils/zcsvgutils';
import { validateNullVal, mergeJSON, getValByPriority, pick, utils } from '../utils/zcutils';
import { mathMin } from '../core/zc-constants';
import { NULL } from '../core/zc-constants';
import { $Browser } from '../core/browser-detect';
import { getDashArray, getLineCap } from '../utils/dashstyles';
import { effects, getGradientProp } from '../effects/zceffects';
import { colorscale_helpers } from '../axis/colorscale-helpers';

var hexToRGB = legend_helpers.hexToRGB;
function getChartTypeSymbol(chartName) {
    return function(size) {
        switch (chartName) {
            case constants.BAR_CHARTNAME:
            case constants.GANTT_CHARTNAME:
            case constants.BULLET_CHARTNAME:
            case constants.WATERFALL_CHARTNAME:
                return (
                    `M${-size * 0.5} ${size * 0.5} L${-size * 0.5} ${-size * 0.2} L${-size * 0.25} ${-size * 0.2} L${-size * 0.25} ${size *
                        0.5}Z ` +
                    ` M${-size * 0.125} ${size * 0.5} L${-size * 0.125} ${-size * 0.5} L${size * 0.125} ${-size * 0.5} L${size *
                        0.125} ${size * 0.5}Z ` +
                    ` M${size * 0.25} ${size * 0.5} L${size * 0.25} ${size * 0.1} L${size * 0.5} ${size * 0.1} L${size * 0.5} ${size *
                        0.5}Z`
                );
            case constants.AREA_CHARTNAME:
            case constants.AREARANGE_CHARTNAME:
                return `M${-size * 0.5} ${size * 0.5} L${-size * 0.5} ${size * 0.28} L${-size * 0.24} ${size * 0.08} L${-size *
                    0.13} ${size * 0.19} L${size * 0.07} ${-size * 0.14} L${size * 0.23} ${size * 0.012} L${size * 0.5} ${-size *
                    0.5} L${size * 0.5} ${size * 0.5}Z`;
            case 'areaLine':
                return `M${-size * 0.5} ${size * 0.28} L${-size * 0.24} ${size * 0.08} L${-size * 0.13} ${size * 0.19} L${size *
                    0.07} ${-size * 0.14} L${size * 0.23} ${size * 0.012} L${size * 0.5} ${-size * 0.5}`;
            case constants.PIE_CHARTNAME:
            case constants.BUBBLEPIE_CHARTNAME:
            case constants.SUNBURST_CHARTNAME:
            case constants.GEOBUBBLEPIE_CHARTNAME:
                return (
                    `M${size * 0.045} ${-size * 0.045} L${size * 0.5} ${-size * 0.045} A ${size * 0.455} ${size * 0.455} 0 0 0 ${size *
                        0.045} ${-size * 0.5}Z` +
                    ` M${-size * 0.045} ${size * 0.045} L${-size * 0.045} ${-size * 0.41} A ${size * 0.455} ${size * 0.455} 1 1 0 ${size *
                        0.41} ${size * 0.045}Z`
                );
            case constants.FUNNEL_CHARTNAME:
                return `M${-size * 0.5} ${-size * 0.5} L ${size * 0.5} ${-size * 0.5} L ${size * 0.15} ${size * 0.2} L ${size *
                    0.15} ${size * 0.5} L ${-size * 0.15} ${size * 0.5} L ${-size * 0.15} ${size * 0.2}Z`;
            case constants.PYRAMID_CHARTNAME:
                return (
                    `M0 ${-size * 0.5} L${size * 0.23} ${-size * 0.06} L${-size * 0.23} ${-size * 0.06}Z` +
                    ` M${-size * 0.5} ${size * 0.5} L${size * 0.5} ${size * 0.5} L${size * 0.27} 0` +
                    ` L${-size * 0.27} 0Z`
                );
            case constants.WEB_CHARTNAME:
                return getSVGsymbol('line')(size);
            case constants.TREEMAP_CHARTNAME:
                return getSVGsymbol('square')(size);
            case constants.SCATTER_CHARTNAME:
            case constants.GEOSCATTER_CHARTNAME:
            case constants.BUBBLE_CHARTNAME:
            case constants.GEOBUBBLE_CHARTNAME:
            case constants.PACKEDBUBBLE_CHARTNAME:
            case constants.LINE_CHARTNAME:
            default:
                return getSVGsymbol(chartName)(size);
        }
    };
}

legend.getHTML4ChartTypeSymbol = function(width, height, fillColor, chartObj, index, disabled) {
    var isQuantileColorScale = colorscale_helpers.isQuantileColorScale(chartObj),
        seriesIndex =
            (chartObj.isAxisCategory || chartObj.dataObject.isPolarAxisCategory || chartObj.dataObject.isHierarchy || chartObj.isGeoMap) &&
            !isQuantileColorScale
                ? index
                : 0,
        chartName = chartObj.chartTypes[seriesIndex].name,
        commonPlotOptions = chartObj.systemConf.chart.plot.plotoptions[chartName],
        plotOptions = isQuantileColorScale ? NULL : chartObj.seriesdata[seriesIndex].plotoptions,
        symbolpath = getChartTypeSymbol(chartName)(width),
        gvp = utils.getValueOfPlotOptions(plotOptions, commonPlotOptions);

    switch (chartName) {
        case constants.AREA_CHARTNAME:
        case constants.AREARANGE_CHARTNAME: {
            let strokeWidth = mathMin(pick(gvp('strokeWidth'), 2), height * 0.15),
                gradientProperties = getGradientProperties(plotOptions, commonPlotOptions, width, height, fillColor, chartObj, chartName),
                dashProperties = getDashProperties(plotOptions, commonPlotOptions, strokeWidth),
                markerOptions =
                    plotOptions && plotOptions.marker
                        ? mergeJSON(commonPlotOptions.marker, plotOptions.marker, true)
                        : commonPlotOptions.marker,
                //
                fill = disabled ? fillColor : gradientProperties.fill || gvp('fillColor') || fillColor,
                stroke = disabled ? fillColor : gvp('strokeColor') || fillColor,
                fillOpacity = pick(gvp('fillOpacity'), 1),
                strokeOpacity = gvp('strokeOpacity'),
                //
                markerProperties = '',
                areaStyles = `fill:${hexToRGB(fill)};fill-opacity:${fillOpacity};`,
                lineStyles = `stroke:${hexToRGB(stroke)};stroke-width:${strokeWidth};stroke-opacity:${strokeOpacity};fill:none;${
                    dashProperties ? dashProperties : ''
                }`;

            if (markerOptions.enabled != false) {
                let outerRatio = 0.5,
                    innerRatio = 0.35;

                markerProperties = getMarkerProperties(
                    markerOptions,
                    width,
                    height,
                    fillColor,
                    'outer',
                    mathMin(width, height) * outerRatio
                );
                markerProperties += getMarkerProperties(
                    markerOptions,
                    width,
                    height,
                    fillColor,
                    'inner',
                    mathMin(width, height) * innerRatio
                );
            }
            let areaProperties = `<path d='${symbolpath}' transform='translate(${width / 2},${height / 2})' style='${areaStyles}'></path>`,
                lineProperties = `<path d='${getChartTypeSymbol('areaLine')(width)}' transform='translate(${width / 2},${height /
                    2})' style='${lineStyles}'></path>`;
            return `${gradientProperties.defs}` + areaProperties + lineProperties + markerProperties;
        }
        case constants.LINE_CHARTNAME:
        case constants.WEB_CHARTNAME:
            var strokeWidth = mathMin(pick(gvp('strokeWidth'), 2), height * 0.15),
                markerOptions =
                    plotOptions && plotOptions.marker
                        ? mergeJSON(commonPlotOptions.marker, plotOptions.marker, true)
                        : commonPlotOptions.marker,
                stroke = disabled ? fillColor : gvp('strokeColor') || fillColor,
                dashProperties = getDashProperties(plotOptions, commonPlotOptions, strokeWidth),
                markerProperties = '';
            var styles = `stroke:${hexToRGB(stroke)};` + `stroke-width:${strokeWidth};`;

            if (dashProperties) {
                styles += dashProperties;
            } else if (markerOptions.enabled != false) {
                var outerRatio = chartName === constants.WEB_CHARTNAME ? 0.5 : 0.65,
                    innerRatio = chartName === constants.WEB_CHARTNAME ? 0.25 : 0.45;

                markerProperties = getMarkerProperties(
                    markerOptions,
                    width,
                    height,
                    fillColor,
                    'outer',
                    mathMin(width, height) * outerRatio
                );
                markerProperties += getMarkerProperties(
                    markerOptions,
                    width,
                    height,
                    fillColor,
                    'inner',
                    mathMin(width, height) * innerRatio
                );
            }

            return `<path d='${symbolpath}' transform='translate(${width / 2},${height / 2})' style='${styles}'></path>${markerProperties}`;

        case constants.SCATTER_CHARTNAME:
        case constants.BUBBLE_CHARTNAME:
        case constants.GEOBUBBLE_CHARTNAME:
        case constants.GEOSCATTER_CHARTNAME:
            var markerOptions =
                plotOptions && plotOptions.marker
                    ? mergeJSON(commonPlotOptions.marker, plotOptions.marker, true)
                    : commonPlotOptions.marker;

            if (markerOptions.imageUrl) {
                return markerOptions.imageUrl;
            }

            var outerSize = mathMin(width, height) - validateNullVal(markerOptions.outerStrokeWidth, 1),
                innerSize = mathMin(width, height) * 0.5 - validateNullVal(markerOptions.innerStrokeWidth, 1),
                outerGradientProperties = getGradientProperties(
                    plotOptions,
                    commonPlotOptions,
                    outerSize,
                    outerSize,
                    markerOptions.outerFillColor || fillColor,
                    chartObj,
                    chartName
                ),
                innerGradientProperties = getGradientProperties(
                    plotOptions,
                    commonPlotOptions,
                    innerSize,
                    innerSize,
                    markerOptions.innerFillColor || fillColor,
                    chartObj,
                    chartName
                );

            return (
                outerGradientProperties.defs +
                getMarkerProperties(markerOptions, width, height, fillColor, 'outer', outerSize, outerGradientProperties.fill) +
                innerGradientProperties.defs +
                getMarkerProperties(markerOptions, width, height, fillColor, 'inner', innerSize, innerGradientProperties.fill)
            );
        case constants.PACKEDBUBBLE_CHARTNAME:
            var size = mathMin(width, height) - validateNullVal(commonPlotOptions.strokeWidth, 1),
                gradientProperties = getGradientProperties(plotOptions, commonPlotOptions, size, size, fillColor, chartObj, chartName);
            return (
                gradientProperties.defs +
                getMarkerProperties(commonPlotOptions, width, height, fillColor, null, size, gradientProperties.fill)
            );
        case constants.HEATMAP_CHARTNAME:
            if (commonPlotOptions.symbol && getSVGsymbol(commonPlotOptions.symbol)) {
                var size = mathMin(width, height) - validateNullVal(commonPlotOptions.strokeWidth, 1),
                    gradientProperties = getGradientProperties(plotOptions, commonPlotOptions, size, size, fillColor, chartObj, chartName);
                return (
                    gradientProperties.defs +
                    getMarkerProperties(commonPlotOptions, width, height, fillColor, null, size, gradientProperties.fill)
                );
            } else {
                var strokeWidth = validateNullVal(commonPlotOptions.strokeWidth, 1),
                    w = width - strokeWidth,
                    h = height - strokeWidth,
                    gradientProperties = getGradientProperties(plotOptions, commonPlotOptions, w, h, fillColor, chartObj, chartName);
                return (
                    gradientProperties.defs +
                    getRoundedRectProperties(commonPlotOptions, w, h, fillColor, chartObj, gradientProperties.fill)
                );
            }
        default: {
            let gradientProperties = getGradientProperties(plotOptions, commonPlotOptions, width, height, fillColor, chartObj, chartName),
                fillOpacity = pick(plotOptions && plotOptions.fillOpacity, commonPlotOptions.fillOpacity, 1),
                styles = `fill:${gradientProperties.fill || fillColor};` + `fill-opacity:${fillOpacity};`; //Need to support stroke properties????
            return `${gradientProperties.defs}<path d='${symbolpath}' transform='translate(${width / 2},${height /
                2})' style='${styles}'></path>`;
        }
    }
};

function getDashProperties(plotOptions, commonPlotOptions, strokeWidth) {
    var dashStyle = getValByPriority([plotOptions, commonPlotOptions], 'dashStyle'),
        dashArray = getDashArray(dashStyle, strokeWidth * 0.75),
        styles;

    if (dashArray && dashArray.length) {
        var strokeLineCap = getLineCap(getValByPriority([plotOptions, commonPlotOptions], 'lineCap'), dashStyle);
        styles = `stroke-dasharray:${dashArray};` + `stroke-linecap:${strokeLineCap};`;
    }

    return styles;
}

function getMarkerProperties(markerOptions, width, height, color, markerType, size, gradientFill) {
    var symbolpath = getChartTypeSymbol(markerOptions.symbol)(size);

    var styles =
        `stroke:${hexToRGB(markerOptions[markerType ? `${markerType}StrokeColor` : 'strokeColor'] || color)};` +
        `stroke-width:${validateNullVal(markerOptions[markerType ? `${markerType}StrokeWidth` : 'strokeWidth'], 1)};` +
        `stroke-opacity:${validateNullVal(markerOptions[markerType ? `${markerType}StrokeOpacity` : 'strokeOpacity'], 1)};` +
        `fill:${gradientFill || hexToRGB(markerOptions[markerType ? `${markerType}FillColor` : 'fillColor'] || color)};` +
        `fill-opacity:${validateNullVal(markerOptions[markerType ? `${markerType}FillOpacity` : 'fillOpacity'], 1)};`;

    return `<path d='${symbolpath}' transform='translate(${width / 2},${height / 2})' style='${styles}'></path>`;
}

function getGradientProperties(plotOptions, commonPlotOptions, width, height, color, chartObj, chartName) {
    var gradientOptions =
            plotOptions && plotOptions.gradients
                ? mergeJSON(commonPlotOptions.gradients || {}, plotOptions.gradients, true)
                : commonPlotOptions.gradients || {},
        defs = '',
        fill,
        patternSvg = chartObj.getPatternSvg();

    if (gradientOptions.type === 'linear' || gradientOptions.type === 'radial') {
        var prop = getGradientProp(gradientOptions, [width, height], color, 1),
            args = {
                offset: chartName,
                chartObj: chartObj
            },
            gradientID = effects.get(effects.svg_gradients, 'svgGradientFill', prop, args),
            XMLS = new XMLSerializer(),
            // #ZC1530
            h = XMLS.serializeToString(patternSvg.select(`#${gradientID}`).node()); // Converting DOM node into a string

        h = h.replace(/"/g, "'");
        defs = `<defs>${h}</defs>`;
        fill = `url(${$Browser.IE ? '#' : '%23'}${gradientID})`; //https://www.chromestatus.com/features/5656049583390720  (Using unescaped '#' characters in a data URI body is deprecated and will be removed in M71, around December 2018)
    }

    return {
        defs: defs,
        fill: fill
    };
}

function getRoundedRectProperties(commonPlotOptions, width, height, color, chartObj, gradientFill) {
    var borderRadius = commonPlotOptions.borderRadius;
    if (borderRadius) {
        borderRadius = mathMin(mathMin(width, height) / 4, borderRadius);
    }

    var path = roundedRect(
        -width / 2,
        -height / 2,
        width,
        height,
        chartObj.dataObject.getTransRectValue('radius', null, null, null, null, borderRadius)
    );

    var styles =
        `stroke:${hexToRGB(commonPlotOptions.strokeColor || color)};` +
        `stroke-width:${validateNullVal(commonPlotOptions.strokeWidth, 1)};` +
        `stroke-opacity:${validateNullVal(commonPlotOptions.strokeOpacity, 1)};` +
        `fill:${gradientFill || hexToRGB(commonPlotOptions.fillColor || color)};` +
        `fill-opacity:${validateNullVal(commonPlotOptions.fillOpacity, 1)};`;

    return `<path d='${path}' transform='translate(${width / 2},${height / 2})' style='${styles}'></path>`;
}

export var _legendSymbols = legend.getHTML4ChartTypeSymbol;
