// @zohocharts/zc/src/legend/legend-highlight.js
//$Id$
//legend-highlight.js

import { DataProcessor } from '../core/data-processor';
import { Registry } from '../core/zc-registry';
// import { legend_helpers } from './legend-helper';
import { charttypenames, charttype, chartcategory, renderer_helpers, plot, UNHIGHLIGHT_EFFECT_REGEX, legend_helpers as L_H } from '../core/main-zc-init';
import {
    invokeFunction,
    defined,
    utils,
    simpleClone,
    arrayIncludes,
    splat,
    isArray,
    getPropVal,
    getVal,
    pick,
    isUndefined
} from '../utils/zcutils';
import { stringifiedArrayReverse, stringIncludes, isObjectPropDefined } from '../utils/zcutils';
import { effects } from '../effects/zceffects';
import { getColorCopy, getColorByIndex, rgba_canvas, getGradients, getRawColor, getHoverGradients } from '../utils/colorutils';
import { $Browser } from '../core/browser-detect';
import { NONE, NULL } from '../core/zc-constants';
import { NUMERIC_DATATYPE } from '../core/datatype-constants';
import { d3_select, d3_selectAll, d3_hsl } from '../core/d3-constants';
import { isFunction } from '../utils/function-utils';
import { addPatternThemeByEle, patternThemes, getPattern } from '../effects/patternThemes';
import { colorscale_helpers } from '../axis/colorscale-helpers';
import {
    isBubbleChartType,
    isBarChartType,
    isScatterChartType,
    isPieChartType,
    isHeatMapChartType,
    isGeoHeatMapChartType
} from '../events/events-utils';
import { isAreaChartType, isCategoryOf } from '../core/charttype-helpers';
import { elementPriority } from './legend-helper';
// import { createSVGRenderer, drawSVG } from '../plot/svgRenderer';
import { SANKEY_CHARTNAME, PIE_CHARTNAME, FUNNEL_CHARTNAME, PYRAMID_CHARTNAME } from '../core/chartname-constants';
import { AREA_CHARTNAME, AREARANGE_CHARTNAME, WEB_CHARTNAME } from '../core/chartname-constants';
import { isFacet } from '../axis/hierarchy-axis-utils';
import { getGridList } from '../core/data-processor/pivot-layout-helpers';
import { getChartTypeByIndex } from '../core/charttype-helpers';
import { sankeyGradientDimension } from '../sankey/sankey-chart';

export var legendHighlight = function(chartObj, container, options = { encode: 'clr' }) {
    var legendHighlight = {},
        origProp = {},
        defsLocation = chartObj.defsLocation,
        userdata = chartObj.userdata,
        hasColorScale = !chartObj.nochartdata && colorscale_helpers.isQuantileColorScale(chartObj),
        // Options
        { encode, order = 0 } = options,
        // data processor helpers

        dataProcessorHelpers = DataProcessor.helpers,
        getChartType = dataProcessorHelpers.getChartType,
        filterSelectionByColor = L_H.filterSelectionByColor,
        isItemDisabled = L_H.isItemDisabled,
        filterSelection = (selection, selectedSeries, filterIndex, seriesIndex) => {
            let chartType = getChartTypeByIndex(chartObj, '0_0', seriesIndex).value,
                hasSizeLegend = chartcategory.zaxis.indexOf(chartType) > -1 && !isHeatMapChartType(chartType);

            if (encode === 'z' && !hasSizeLegend) {
                return selectedSeries ? d3_selectAll() : selection;
            }

            return filterSelectionByColor(selection, selectedSeries, filterIndex, chartObj, encode);
        },
        config = options.config || chartObj.systemConf.legend;

    /*
     * Highlight effects for Bar,pie,funnel,web,area,line,scatter
     * patterns--> diagonalfill,strokeEnlarge,fadeout
     * diagonal fill not supported for line,scatter
     * fadeout: fillopacity except line and scatter, strokeopacity for line and scatter
     */
    legendHighlight.highlightSelectedSeries = function(legdata, index, filterIndex, isSelection) {
        legendHighlight.baseClr = getCurrectColor(index, filterIndex);

        if (chartObj.isGeoMap && isGeoHeatMapChartType(getChartTypeByIndex(chartObj, '0_0', index).value)) {
            highlightGeoMap(chartObj, legdata, index, filterIndex);
        } else {
            switch (chartObj.plot.renderer.mode) {
                case 'svg': {
                    highlightSVGSeries(legdata, index, filterIndex, isSelection);
                    break;
                }
                case 'canvas': {
                    highlightCanvasSeries(chartObj, legdata, index, filterIndex, isSelection);
                    break;
                }
                default: {
                    highlightNonSVGSeries(legdata, index);
                }
            }
        }
    };

    legendHighlight.highlightUnselectedSeries = function(legdata, index, filterIndex, isSelection) {
        if (chartObj.isGeoMap || chartObj.plot.renderer.mode === 'svg') {
            unHighlightSVGSeries(legdata, index, filterIndex, isSelection);
        } else {
            unHighlightCanvasSeries(legdata, index, filterIndex, isSelection);
        }
    };

    legendHighlight.reset = function(legdata, index, filterIndex, resetOriginalProp) {
        if (chartObj.isGeoMap && isGeoHeatMapChartType(getChartTypeByIndex(chartObj, '0_0', index).value)) {
            resetGeoMapSeries();
        } else {
            index = splat(index);
            switch (chartObj.plot.renderer.mode) {
                case 'svg': {
                    resetSelectedSeries(legdata, index, filterIndex);
                    resetUnselectedSeries(legdata, index, filterIndex);
                    break;
                }
                case 'canvas': {
                    resetSelectedCanvasSeries();
                    resetUnselectedCanvasSeries();

                    break;
                }
            }
        }

        if (resetOriginalProp !== false) {
            //Test_662.json
            origProp = {}; //Deleting the stored style values, will help you to know whether the element is already highlighted are not.
            this.resetSelection();
        }
    };

    legendHighlight.getLegendHighlightEffect = function(index, seriesType, plotOrLegend, paneID) {
        let chartTypeObj = getChartTypeByIndex(chartObj, paneID, index),
            chartName = chartTypeObj.name,
            cType = chartTypeObj.value,
            isAxisCategory = isCategoryOf(chartObj, cType, 'axis'),
            isNonAxisMultiSeriesCategory = isCategoryOf(chartObj, cType, 'nonAxisMultiSeries');

        var plotOptions =
                isAxisCategory || isNonAxisMultiSeriesCategory || chartObj.isGeoMap
                    ? (chartObj.seriesdata[index] || chartObj.seriesdata[0]).plotoptions
                    : chartObj.seriesdata[0].plotoptions,
            commonPlotOptions = chartObj.systemConf.chart.plot.plotoptions[chartName],
            legendEffect = config.highlightEffect[seriesType],
            ef =
                plotOptions && plotOptions.legendHighlightEffect
                    ? utils.getValueOfPlotOptions(plotOptions.legendHighlightEffect, commonPlotOptions.legendHighlightEffect)(seriesType)
                    : commonPlotOptions.legendHighlightEffect && commonPlotOptions.legendHighlightEffect[seriesType];
            
            if(seriesType === 'selection'){//need to developed 
                var plot = chartObj.systemConf.chart.plot.eventHandler;
                ef = plot && plot.click.behaviour.rejectionSet ? 'customRejectionEffect' : 'defaultRejectionEffect';
            }

        if (plotOrLegend === 'plot') {
            if (seriesType === 'unselectedSeries' && !config.highlightEffect.applyUnselectedSeriesToPlot) {
                return NONE;
            } else {
                return ef || legendEffect || defaultEffect(chartName);
            }
        } else {
            var defaultEf =
                seriesType === 'selectedSeries' && hasColorScale && (isAreaChartType(cType) || cType === charttype.line)
                    ? 'strokeEnlarge(6)'
                    :  defaultEffect(chartName);

            return ef || legendEffect || defaultEf;
        }
    };

    legendHighlight.getEffect = function(effect, chartName, baseClr, additionalStyles, index, from, point, originalChartName, element) {
        if (chartName === 'waterfall') {
            baseClr = (d, i) => getRawColor(chartObj, i, d, index);
        } else {
            let customColor = getUserFillColor(index, originalChartName || chartName); // originalChartName: map-evens sends chartName as 'area' always
            baseClr = customColor || baseClr || legendHighlight.baseClr;
        }

        let isColorFunction = isFunction(baseClr);
        var userEffect = invokeFunction(effect, chartObj, baseClr, patternThemes, from || 'legend', element, point);
        if (userEffect && userEffect.constructor == Object) {
            return userEffect;
        }
        effect = effect === null || invalidEffect(effect, chartName) || userEffect === null ? defaultEffect(chartName) : effect;
        var eStyle = {};

        if (chartObj.patternPallete && /(flip|moody|illuminate)/.test(effect)) {
            var svgPattern = chartObj.patternPallete[(index || 0) % chartObj.patternPallete.length];
            var prop = patternThemes[svgPattern] ? getPattern(svgPattern) : addPatternThemeByEle(svgPattern);

            var effectCallBack = function(data, index) {
                var baseColor = getVal(baseClr, [data, index]);
                if (prop.ele) {
                    // If custom pallete is given, moody effect will be applied
                    prop.bgClr_base = getColorCopy(baseColor, 1, 'darker', d3_hsl);
                } else {
                    var colorCopy =
                        effect === 'flip' ? baseColor : getColorCopy(baseColor, 1, effect === 'moody' ? 'darker' : 'brighter', d3_hsl);
                    prop.bgClr_base = colorCopy;
                    prop.stroke_base = colorCopy;
                    prop.customStaticClr = chartObj.userdata.legend.customPatternColor;

                    if (effect === 'flip') {
                        prop = simpleClone(prop); // To avoid inplace change in the pattern object

                        //swapping bgClr_base and stroke_base
                        var temp = prop.bgClr_base;
                        prop.bgClr_base = prop.stroke_base;
                        prop.stroke_base = temp;

                        //swapping bgClr and stroke
                        var copy_bgClr = stringifiedArrayReverse(prop.bgClr),
                            copy_stroke = stringifiedArrayReverse(prop.stroke);

                        prop.bgClr = copy_stroke;
                        prop.stroke = copy_bgClr;
                    }
                }

                var patternId = effects.get(effects.svg_patterns, 'svgPatternFill', prop, { chartObj: chartObj });
                return `url(${defsLocation}#${patternId})`;
            };
            eStyle.fill = isColorFunction ? effectCallBack : effectCallBack();
        } else if (/(saturate|grayscale|blur|glow|shadow|boxShadow)/.test(effect)) {
            var filterType = /(saturate|grayscale)/.test(effect) ? 'colorMatrix' : 'dropShadow',
                options;
            switch (effect) {
                case 'saturate':
                    options = {
                        type: 'saturate',
                        values: '0.3'
                    };
                    break;
                case 'grayscale':
                    options = {
                        type: 'matrix',
                        values: '.33 .33 .33 0 0 .33 .33 .33 0 0 .33 .33 .33 0 0 0 0 0 1 0'
                    };
                    break;
                case 'blur':
                    options = {
                        blur: 2,
                        source: 'SourceGraphic'
                    };
                    eStyle.opacity = 0.7;
                    break;
                case 'glow':
                    options = {
                        blur: 3,
                        color: 'gold'
                    };
                    break;
                case 'shadow':
                    options = {
                        dx: 5,
                        dy: 5,
                        blur: 3
                    };
                    break;
                case 'boxShadow':
                    eStyle.fill = baseClr;
                    eStyle.stroke = 'white';
                    options = {
                        dx: 0,
                        dy: 0,
                        blur: 5,
                        color: 'rgba(0,0,0,0.3)'
                    };
                    break;
            }
            var filterId = effects.get(effects.svg_filters, filterType, options, { chartObj: chartObj });
            eStyle.filter = `url(${defsLocation}#${filterId})`;
        } else if (/(moody|illuminate)/.test(effect)) {
            var ismoody = effect === 'moody';
            var effectCallback = function(data, index) {
                var baseColor = getVal(baseClr, [data, index]);

                if (stringIncludes(baseColor, 'url')) {
                    return ismoody ? 'grey' : 'silver';
                } else {
                    if (isArray(baseColor)) {
                        return baseColor.map(color => getColorCopy(color, 0.6, ismoody ? 'darker' : 'brighter', d3_hsl));
                    } else {
                        return getColorCopy(baseColor, 0.6, ismoody ? 'darker' : 'brighter', d3_hsl);
                    }
                }

                // return baseColor.indexOf('url') != -1
                //     ? ismoody
                //         ? 'grey'
                //         : 'silver'
                //     : getColorCopy(baseColor, 0.6, ismoody ? 'darker' : 'brighter', d3_hsl);
            };
            eStyle.fill = isColorFunction ? effectCallback : effectCallback();
        } else if (effect.indexOf('strokeEnlarge') > -1) {
            var customStrokeWidth = effect.split(/(\d+)/g)[1];
            eStyle['stroke-width'] = defined(customStrokeWidth) ? +customStrokeWidth : 3;

            if (hasColorScale) {
                eStyle.visibility = NULL;
            }

            if (additionalStyles) {
                for (var a in additionalStyles) {
                    eStyle[a] = additionalStyles[a];
                }
            }
        } else if (effect === 'fadeOut') {
            eStyle.opacity = 0.3;
        } else if (effect === 'blackyBorder') {
            eStyle.stroke = 'rgba(0,0,0,0.5)';
            eStyle['stroke-width'] = 1;
        } else if (effect === 'invert') {
            eStyle.fill = 'white';
            eStyle.stroke = baseClr;
            eStyle['stroke-width'] = 1;
            eStyle['paint-order'] = null;
        } else if (effect === 'ridge') {
            eStyle.fill = baseClr;
            eStyle.stroke = 'white';
            eStyle['stroke-width'] = 7;
        } else if (effect !== NONE) {
            var prop = getPattern(effect);

            if (prop) {
                var effectCallback = function(data, index) {
                    var baseColor = getVal(baseClr, [data, index]);
                    var prop = getPattern(effect);
                    prop.bgClr_base = baseColor;
                    prop.stroke = 'transparent,white,white'; //need to check
                    prop.gradient = additionalStyles && additionalStyles.gradient;
                    prop.bound = additionalStyles && additionalStyles.bound;
                    const submodule = prop.gradient ? 'svgPatternWithGradientFill' : 'svgPatternFill';
                    var patternId = effects.get(effects.svg_patterns, submodule, prop, { chartObj: chartObj });
                    return `url(${defsLocation}#${patternId})`;
                };
                eStyle.fill = isColorFunction ? effectCallback : effectCallback();
            } else {
                eStyle.fill = baseClr;
            }
        }

        // check if the gradient is enabled
        let optionsKey = chartName;
        // #ZC2586 $NONE
        if ((chartName === 'sankey' || chartName === 36) && effect !== NONE) {
            optionsKey = `${optionsKey}.link`;
            // check the hover options

            // #ZC2584 chartName is passed as number. we should pass as chart name

            // can be done for all charges
            let hoverOptions = getPropVal(chartObj.systemConf, `chart.plot.plotoptions.sankey.hoveroptions`);
            eStyle['fill-opacity'] = hoverOptions.fillOpacity;
        }

        let gradientOptions = getPropVal(chartObj.systemConf, `chart.plot.plotoptions.${optionsKey}.gradients`) || {};

        if (defined(gradientOptions.type) && isArray(baseClr) && chartName === SANKEY_CHARTNAME) {
            gradientOptions.options[gradientOptions.type].stopColor = baseClr;
            if (!gradientOptions.customGradient && point) {
                sankeyGradientDimension(point, gradientOptions);
            }

            let gradientDimension = [chartObj.plotarea.width, chartObj.plotarea.height];
            eStyle.fill = getGradients(gradientOptions, gradientDimension, baseClr, 'svg', 1, chartName, false, chartObj);
        }

        return eStyle;
    };

    legendHighlight.raiseSelection = function(elements, paneID) {
        if (doArrangeSelection(paneID)) {
            elements.raise();
        }
    };

    legendHighlight.resetSelection = function() {
        let seriesHolderWrap = container.selectAll('g.seriesholder');

        if (isFacet(chartObj)) {
            var gridList = getGridList(chartObj.dataObject);

            gridList.forEach(grid => {
                let { row, column } = grid,
                    paneID = `${row}_${column}`,
                    gridWrap = seriesHolderWrap.selectAll(`g.grid${paneID}`);

                if (doArrangeSelection(paneID)) {
                    _resetSelection(gridWrap);
                }
            });
        } else {
            if (doArrangeSelection()) {
                _resetSelection(seriesHolderWrap);
            }
        }
        seriesHolderWrap = null;
    };

    function _resetSelection(g) {
        var elements;

        if (chartObj.dataObject.noofAreaEnabledSeries || chartObj.dataObject.noofAreaRangeEnabledSeries) {
            var getAllChildNodes = function() {
                    return this.childNodes;
                },
                getAllGroupNodes = function() {
                    return this.tagName == 'g';
                };
            elements = g.selectAll(getAllChildNodes).filter(getAllGroupNodes);
        } else if (chartObj.dataObject.isFlowChartCategory) {
            elements = g.selectAll('g.sankey-links').selectAll('path.sankey-link');
        } else {
            elements = g.selectAll('g.hierarchychart').selectAll('*');
        }

        // elements.sort((a, b) => a.index - b.index);
        elements.sort((a, b) => {
            if (a.stackIndex !== undefined && b.stackIndex !== undefined) {
                return (a.stackIndex <= b.stackIndex) ? -1 : 1;
            } else {
                return a.index - b.index;
            }
        })
    }

    function resetSelectedSeries(legdata, index, filterIndex) {
        legendHighlight.selSeriesObj = [];
        index.forEach(function(ii) {
            var selected = getSelectedSeries(ii, filterIndex),
                selecteffect = legendHighlight.getLegendHighlightEffect(ii, 'selectedSeries'),
                selectedSeries = selected.selectedSeries,
                charttypes = selected.charttypes,
                paneIDs = selected.paneIDs,
                length = selectedSeries.length;

            for (var i = 0; i < length; i++) {
                let prop,
                    cType = charttypenames.get(charttypes[i]);

                if (chartObj.dataObject.isFlowChartCategory) {
                    // eslint-disable-next-line no-loop-func
                    selectedSeries[i].each(function(p) {
                        // every sankey point has different colors. so getting the color for individuals points
                        prop = legendHighlight.getEffect(selecteffect, cType, null, null, ii);
                        resetProp(d3_select(this), prop, p.point.itemIndex, 'select');
                    });
                } else {
                    prop = legendHighlight.getEffect(selecteffect, cType, null, null, ii);
                    resetProp(selectedSeries[i], prop, ii, 'select', paneIDs[i]);
                }
            }
            legendHighlight.selSeriesObj.push(selected);
        });
    }

    function resetUnselectedSeries(legdata, index, filterIndex) {
        legendHighlight.unselSeriesObj = getUnselectedSeries(index, filterIndex);
        var unselectedSeries = legendHighlight.unselSeriesObj.unselectedSeries;
        var indices = legendHighlight.unselSeriesObj.indices;
        var charttypes = legendHighlight.unselSeriesObj.charttypes;
        var paneIDs = legendHighlight.unselSeriesObj.paneIDs;
        var length = unselectedSeries.length;
        for (var i = 0; i < length; i++) {
            var cType = charttypenames.get(charttypes[i]);
            var unselecteffect = legendHighlight.getLegendHighlightEffect(indices[i], 'unselectedSeries');

            if (chartObj.dataObject.isFlowChartCategory) {
                // eslint-disable-next-line no-loop-func
                unselectedSeries[i].each(function(p) {
                    // every sankey point has different colors. so getting the color for individuals points
                    prop = legendHighlight.getEffect(unselecteffect, cType, NULL, NULL, indices[i]);
                    resetProp(d3_select(this), prop, p.point.itemIndex, 'unselect');
                });
            } else {
                var prop = legendHighlight.getEffect(unselecteffect, cType, null, null, indices[i]);
                resetProp(unselectedSeries[i], prop, indices[i], 'unselect', paneIDs[i]);
            }
        }
    }

    function getSelectedSeries(index, filterIndex, isSelection) {
        var _seriesHolder = container.selectAll('g.seriesholder');
        var type = getChartType(null, 'object', index, userdata).value;
        var charttypes = [],
            selectedSeries = [],
            parentSeries = [],
            paneIDs = [];

        if (isFacet(chartObj)) {
            if (noColorColumn(chartObj)) {
                collectSelectedSeries(type, index, _seriesHolder);
            } else {
                var gridList = getGridList(chartObj.dataObject),
                    count = -1;

                gridList.forEach(grid => {
                    let { row, column, layer } = grid,
                        paneid = `${row}_${column}`,
                        gridWrap = _seriesHolder.selectAll(`g.grid${paneid}`);

                    grid.data.forEach((_d, _i) => {
                        let enabled = !isItemDisabled(_d);
                        count++;
                        if (
                            (!chartObj.userdata.facets.color.length || layer === order) &&
                            chartObj.rendererConf[count].legendIndex() === index &&
                            enabled
                        ) {
                            type = getChartTypeByIndex(chartObj, paneid, _i).value;
                            collectSelectedSeries(type, _i, gridWrap, paneid);
                        }
                    });
                });
            }
        } else {
            collectSelectedSeries(type, index, _seriesHolder, 0, isSelection);
        }

        return {
            selectedSeries: selectedSeries,
            charttypes: charttypes,
            parentSeries: parentSeries,
            paneIDs
        };

        function collectSelectedSeries(type, index, seriesHolder, paneID = 0, isSelection) {
            /*
             * This function adds the selected series to the selectedSeries array and respective chart type to the charttypes array.
             * Scatter chart = outer scatter series + inner scatter series.
             * Area chart = area series + line series.
             * Line chart = line series + scatter series.(for IE and SAFARI)
             */
            var groupHolderIndex = index;
            if (chartObj.seriesAddRemoveDetails) {
                // addRemoveSeriesPhase
                groupHolderIndex = chartObj.renderer[index].commonRendererProp().params.dynamicSeriesPosition;
            }

            var series, parent;
            if (isHeatMapChartType(type)) {
                var heatmapSeries = seriesHolder.selectAll(`g.heatmapseries_${index}`);
                series = heatmapSeries.selectAll('path');
            } else if (isBarChartType(type)) {
                if (chartObj.dataObject.isPolarAxisCategory) {
                    series = seriesHolder.selectAll('[class^=barseries]').selectAll(`path.zc_bar_${index}`);
                } else {
                    var barSeries = seriesHolder.selectAll(`g.barseries_${groupHolderIndex}`);
                    series = barSeries.selectAll('path.zc_bar');
                }

                //whiskers
                selectedSeries.push(chartObj.staticPaneSVG.selectAll(`g.medianSeries${index}`).selectAll('path.outer'));
                selectedSeries.push(chartObj.staticPaneSVG.selectAll(`g.whiskerSeries${index}`).selectAll('path.outer'));
                charttypes.push(charttype.scatter);
                charttypes.push(charttype.scatter);
                paneIDs.push(paneID, paneID);
            } else if (type == charttype.line) {
                var lineSeries = seriesHolder.selectAll(`g.lineseries_${groupHolderIndex}`);
                series = lineSeries.selectAll('path.line');
                parent = lineSeries;
            } else if (isAreaChartType(type) || type == charttype.web) {
                var areaSeries = seriesHolder.selectAll(`g.areaseries_${groupHolderIndex}`);
                series = areaSeries.selectAll('path.area');
                parent = areaSeries;
            } else if (isScatterChartType(type)) {
                var scatterSeries = seriesHolder.selectAll(`g.scatterseries_${groupHolderIndex}`);
                series = scatterSeries.selectAll('g.scatter');
                series = filterSelection(series, true, filterIndex, index);

                var outerScatters = series.selectAll('path.outer');
                var innerScatters = series.selectAll('path.inner');

                selectedSeries.push(outerScatters, innerScatters);
                charttypes.push(type, type);
                paneIDs.push(paneID, paneID);

                parent = scatterSeries;
                parentSeries.push(parent);
                return;
            } else if (type === charttype.bubblepie || type === charttype.geobubblepie) {
                var pieSeries = seriesHolder.selectAll('g.piegroup');
                series = pieSeries.selectAll(`path#arc_${groupHolderIndex}`);
            } else if (isPieChartType(type) || type === charttype.dial) {
                var pieSeries = seriesHolder.selectAll('g.piegroup');
                series = pieSeries.selectAll(defined(filterIndex) ? 'path' : `path#arc_${index}`);
            } else if (type == charttype.funnel) {
                var funnelSeries = seriesHolder.selectAll('g.funnelgroup');
                var trapezoidgroup = funnelSeries.selectAll('g.trapezoidgroup');
                series = trapezoidgroup.selectAll(defined(filterIndex) ? 'path' : `path#funnel${index}`);

                //ellipse
                if(isSelection){
                    var ellipsegroup = funnelSeries.selectAll('g.ellipsegroup');
                    var ellipses = ellipsegroup.selectAll(defined(filterIndex) ? 'path' : `path#ellipse${index}`);
                    selectedSeries.push(ellipses);
                }

            } else if (type == charttype.pyramid) {
                var pyramidSeries = seriesHolder.selectAll('g.pyramidgroup');
                series = pyramidSeries.selectAll(defined(filterIndex) ? 'g' : `g#section_${index}`);
            } else if (type == charttype.wordcloud) {
                var wordCloudSeries = seriesHolder.selectAll('g.wordcloudgroup');
                series = wordCloudSeries.selectAll(`text.series_${groupHolderIndex}`);
            } else if (chartcategory.hierarchy.indexOf(type) != -1) {
                var burstseries = seriesHolder.selectAll('g.hierarchychart');
                series = burstseries.selectAll(`.series_${index}`);
            } else if (type == charttype.sankey) {
                let point = chartObj.cache.sankey.legend[index];
                index = (filterIndex !== undefined) ? filterIndex : index; 
                if (chartObj.systemConf.legend.ranges) {
                    const ranges = chartObj.systemConf.legend.ranges,
                        nodes = chartObj.dataObject.sankeyLayout.completeGraph.nodes;
                    filterIndex = (ranges[index].disabled) ? filterIndex : undefined;
                    let val = Array.isArray(ranges[0]) ? ranges[index][0] : ranges[index];
                    index = nodes.findIndex(d => d.name === val);
                    point = chartObj.cache.sankey.legend[index];
                }
                let dependentsPoints = chartObj.renderer[0].highlighter.getDependentLayoutPoints(point, true, 'legend');
                series = seriesHolder.selectAll('path').filter(currentLayoutPoint => {
                    return arrayIncludes(dependentsPoints, currentLayoutPoint);
                });
            }

            let isAxisCategory = isCategoryOf(chartObj, type, 'axis'),
                isNonAxisMultiSeriesCategory = isCategoryOf(chartObj, type, 'nonAxisMultiSeries'),
                _seriesIndex = isAxisCategory || isNonAxisMultiSeriesCategory || chartObj.isGeoMap ? index : 0;
            series = filterSelection(series, true, filterIndex, _seriesIndex);

            selectedSeries.push(series);
            charttypes.push(type);
            parentSeries.push(parent);
            paneIDs.push(paneID);

            if (isAreaChartType(type) || type == charttype.web) {
                collectSelectedSeries(charttype.line, index, seriesHolder, paneID);
            }
            const isScatterMarkers = !getPropVal(userdata, "chart.plot.plotoptions.area.marker.useSVGMarkers");
            if (type == charttype.line && (hasSubScatterRenderer(filterIndex) || isScatterMarkers)) {
                collectSelectedSeries(charttype.scatter, index, seriesHolder, paneID, isSelection);
            }
        }
    }

    function getUnselectedSeries(index, filterIndex) {
        var unSelectedSeries = [],
            indices = [],
            charttypes = [],
            paneIDs = [],
            parentSeries;
        var _seriesHolder = container.selectAll('g.seriesholder'),
            count = 0;

        if (isFacet(chartObj)) {
            if (noColorColumn(chartObj)) {
                collectUnselectedSeries(null, index, _seriesHolder, chartObj.seriesdata);
            } else {
                var gridList = getGridList(chartObj.dataObject);

                gridList.forEach(grid => {
                    let { row, column, layer } = grid,
                        gridWrap = _seriesHolder.selectAll(`g.grid${row}_${column}`);

                    if (!chartObj.userdata.facets.color.length || layer === order) {
                        collectUnselectedSeries(null, index, gridWrap, grid.data, count, `${row}_${column}`);
                    } else {
                        collectUnselectedSeries(null, [-1], gridWrap, grid.data, count, `${row}_${column}`);
                    }
                    count += grid.data.length;
                });
            }
        } else {
            collectUnselectedSeries(null, index, _seriesHolder, chartObj.seriesdata);
        }

        return {
            unselectedSeries: unSelectedSeries,
            indices: indices,
            charttypes: charttypes,
            parentSeries: parentSeries,
            paneIDs
        };

        function collectUnselectedSeries(subcharttype, index, seriesHolder, seriesdata, startingRendererIndex = 0, paneID = 0) {
            /*
             * This function adds the unselected series to the selectedSeries array, respective chart type to the charttypes array, respective index to the indices array.
             * Scatter chart = outer scatter series + inner scatter series.
             * Area chart = area series + line series.
             * Line chart = line series + scatter series.(for IE and SAFARI)
             */
            let chartType = getChartTypeByIndex(chartObj, paneID, index[0]).value, // Have to check with multiple indices
                isAxisCategory = isCategoryOf(chartObj, chartType, 'axis'),
                isNonAxisMultiSeriesCategory = isCategoryOf(chartObj, chartType, 'nonAxisMultiSeries');

            var series, type;
            if (isAxisCategory || isNonAxisMultiSeriesCategory || chartObj.isGeoMap) {
                seriesdata.forEach(function(d, i) {
                    let legendOrder = noColorColumn(chartObj) ? i : chartObj.rendererConf[startingRendererIndex + i].legendIndex();
                    if (d.disabled != true && (defined(filterIndex) ? arrayIncludes(index, i) : !arrayIncludes(index, legendOrder))) {
                        type = subcharttype || getChartTypeByIndex(chartObj, paneID, i).value;
                        var selectionIndex = defined(filterIndex) ? index : i,
                            groupHolderIndex = selectionIndex;
                        if (chartObj.seriesAddRemoveDetails) {
                            // addRemoveSeriesPhase,
                            groupHolderIndex = chartObj.renderer[i].commonRendererProp().params.dynamicSeriesPosition;
                        }

                        if (isHeatMapChartType(type)) {
                            var heatmapSeries = seriesHolder.selectAll(`g.heatmapseries_${selectionIndex}`);
                            series = heatmapSeries.selectAll('path');
                        } else if (isBarChartType(type)) {
                            var barSeries = seriesHolder.selectAll(`g.barseries_${groupHolderIndex}`);
                            series = barSeries.selectAll('path.zc_bar');

                            var whiskers = chartObj.renderer[selectionIndex].commonRendererProp().params.whiskers,
                                hasWhisker = whiskers && whiskers.enabled && whiskers.showOn == 'overlay';

                            if (hasWhisker) {
                                unSelectedSeries.push(
                                    chartObj.staticPaneSVG.selectAll(`g.whiskerSeries${selectionIndex}`).selectAll('path.outer')
                                );
                                charttypes.push(charttype.scatter);
                                indices.push(selectionIndex, selectionIndex);
                                paneIDs.push(paneID, paneID);
                            }
                        } else if (type == charttype.line) {
                            var lineSeries = seriesHolder.selectAll(`g.lineseries_${groupHolderIndex}`);
                            series = lineSeries.selectAll('path.line');
                        } else if (isAreaChartType(type) || type == charttype.web) {
                            var areaSeries = seriesHolder.selectAll(`g.areaseries_${groupHolderIndex}`);
                            series = areaSeries.selectAll('path.area');
                            parentSeries = areaSeries;
                        } else if (isScatterChartType(type)) {
                            var scatterSeries = seriesHolder.selectAll(`g.scatterseries_${groupHolderIndex}`);
                            series = scatterSeries.selectAll('g.scatter');
                            series = filterSelection(series, false, filterIndex, i);

                            var outerScatters = series.selectAll('path.outer');
                            var innerScatters = series.selectAll('path.inner');

                            unSelectedSeries.push(outerScatters, innerScatters);
                            charttypes.push(type, type);
                            indices.push(selectionIndex, selectionIndex);
                            paneIDs.push(paneID, paneID);
                            return;
                        } else if (chartcategory.hierarchy.indexOf(type) != -1) {
                            var burstseries = seriesHolder.selectAll('g.hierarchychart');
                            series = burstseries.selectAll(`.series_${selectionIndex}`);
                        } else if (type == charttype.wordcloud) {
                            var wordCloudSeries = seriesHolder.selectAll('g.wordcloudgroup');
                            series = wordCloudSeries.selectAll(`.series_${selectionIndex}`);
                        } else if (chartcategory.bubblePie.has(type)) {
                            series = seriesHolder.selectAll('g.piegroup').selectAll(`path#arc_${i}`);
                        }

                        series = filterSelection(series, false, filterIndex, i);

                        unSelectedSeries.push(series);
                        indices.push(i);
                        charttypes.push(type);
                        paneIDs.push(paneID);
                    }
                });
            } else {
                type = subcharttype || chartType;
                var prefix_ID,
                    processFunc = function(d, i) {
                        var ele = d3_select(this),
                            elementIndex = +this.getAttribute('id')
                                .split(prefix_ID)
                                .pop(),
                            unselect = defined(filterIndex)
                                ? !filterSelection(ele, false, filterIndex, 0).empty()
                                : this.getAttribute('id') != prefix_ID + index;

                        if (unselect && (defined(filterIndex) ? true : !arrayIncludes(index, elementIndex))) {
                            unSelectedSeries.push(ele);
                            indices.push(elementIndex);
                            charttypes.push(type);
                            paneIDs.push(paneID);
                        }
                        return unselect;
                    };

                if (isPieChartType(type) || type === charttype.dial) {
                    prefix_ID = 'arc_';
                    series = seriesHolder
                        .selectAll('g.piegroup')
                        .selectAll('path')
                        .filter(processFunc);
                } else if (type == charttype.funnel) {
                    prefix_ID = 'funnel';
                    series = seriesHolder
                        .selectAll('g.trapezoidgroup')
                        .selectAll('path')
                        .filter(processFunc);

                    // ZC2713
                    prefix_ID = 'ellipse';
                    series = seriesHolder
                        .selectAll('g.ellipsegroup')
                        .selectAll('path')
                        .filter(processFunc);
                } else if (type == charttype.pyramid) {
                    prefix_ID = 'section_';
                    var pyramidGroup = seriesHolder.selectAll('g.pyramidgroup');
                    series = pyramidGroup.selectAll('g').filter(processFunc);
                } else if (type === charttype.sankey) {
                    index = (filterIndex !== undefined) ? filterIndex: index; 
                    if (chartObj.systemConf.legend.ranges) {
                        const ranges = chartObj.systemConf.legend.ranges,
                            nodes = chartObj.dataObject.sankeyLayout.completeGraph.nodes;
                        filterIndex = (ranges[index].disabled) ? filterIndex : undefined;
                        let val = Array.isArray(ranges[0]) ? ranges[index][0] : ranges[index];
                        index = nodes.findIndex(d => d.name === val);
                    }
                    let sankeySeriesIndex = 0,
                        point = chartObj.cache.sankey.legend[index],
                        dependentsPoints = chartObj.renderer[sankeySeriesIndex].highlighter.getDependentLayoutPoints(point, true, 'legend');

                    series = seriesHolder.selectAll('path').filter(currentLayoutPoint => {
                        return !arrayIncludes(dependentsPoints, currentLayoutPoint);
                    });

                    unSelectedSeries.push(series);
                    charttypes.push(type);
                    indices.push(sankeySeriesIndex);
                    paneIDs.push(paneID);
                }
            }

            if (isAreaChartType(type) || type == charttype.web) {
                collectUnselectedSeries(charttype.line, index, seriesHolder, seriesdata, startingRendererIndex, paneID);
            }
            if (type == charttype.line && hasSubScatterRenderer(filterIndex)) {
                collectUnselectedSeries(charttype.scatter, index, seriesHolder, seriesdata, startingRendererIndex, paneID);
            }
        }
    }

    function applyProp(series, parentSeries, prop, index, type, paneID) {
        if (prop == null || prop.constructor != Object || series == null || series.node() == null) {
            return;
        }
        var key = getKey(series, index, type, paneID),
            currentOrigProp = {},
            multipleStyles = hasMultipleStyles(index),
            priorityStyles = elementPriority(series);

        for (var attr in prop) {
            if (prop.hasOwnProperty(attr) && !arrayIncludes(priorityStyles, attr)) {
                if (multipleStyles) {
                    currentOrigProp[attr] = currentOrigProp[attr] || [];
                    series.each(function() {
                        currentOrigProp[attr].push(d3_select(this).style(attr));// TODO: fill-opacity gets set to 1 for empty string
                    });
                } else {
                    // TODO Check
                    currentOrigProp[attr] = series.style(attr);
                }

                var val = prop[attr];
                if ($Browser.SAFARI || $Browser.IE) {
                    series
                        .attr(attr, function(d, i) {
                            if(type == 'selection'){
                                return selectionStyling(d, i, val, index,  key, attr, currentOrigProp);
                            }

                            return utils.getVal(val, [d, index]);
                        })
                        .style(attr, function(d, i) {
                            if(type == 'selection'){
                                return selectionStyling(d, i, val, index,  key, attr, currentOrigProp);
                            }

                            return utils.getVal(val, [d, index]);
                        });
                } else {
                        series.style(attr, function(d, i) {
                            if(type == 'selection'){
                                return selectionStyling(d, i, val, index,  key, attr, currentOrigProp);
                            }
                            return utils.getVal(val, [d, index]);
                        });
                }
            }
        }
        origProp[key] = origProp[key] || currentOrigProp; //If the style values are already stored, don't store again. If u store again, the origProp[key] object will be replaced with the highlighting props instead of containing the style values of the original element.

        if (type === 'select') {
            legendHighlight.raiseSelection(parentSeries || series, paneID);
        }
    }

    function selectionStyling(d, i, val, index, key, attr, currentOrigProp){
        var selectionArray = chartObj.cache.clickSelectionData,
            isHierarchyCat = chartObj.dataObject.isHierarchy,
            isAxisCat = chartObj.dataObject.isAxisCategory,
            isFlowChartCat = chartObj.dataObject.isFlowChartCategory;

        if(isFlowChartCat && defined(d.point)){
            d = d.point;
        }
        else if(defined(d.data)){
            d = d.data;
        }

        key = key+"_"+i;//tree map all the children are getting same key
        d.selection = defined(d.selection) ? d.selection : {};
        
        var SelectionGradientObj = d.selection.chartGradient =  defined(d.selection.chartGradient) ? d.selection.chartGradient : {};
            SelectionGradientObj[key] = defined(SelectionGradientObj[key]) ? SelectionGradientObj[key] : {};//to store
            SelectionGradientObj[key][attr] = defined(SelectionGradientObj[key][attr]) ? SelectionGradientObj[key][attr] : (isFlowChartCat ? currentOrigProp[attr] : currentOrigProp[attr][i]);
            if(d.clickSelection == true || isUndefined(selectionArray[0])){//|| chartObj.dataObject.totalnoofdatas == chartObj.cache.selection.selectedPoints.length
                return SelectionGradientObj[key][attr];// to apply oiginal colour from selection/rertain selected element colour
            }

            return utils.getVal(val, [d, index]);
    }

    function resetProp(series, prop, index, type, paneID) {
        if (prop == null || prop.constructor != Object || series == null || series.node() == null) {
            return;
        }

        var key = getKey(series, index, type, paneID),
            currentOrigProp = origProp[key],
            multipleStyles = hasMultipleStyles(index),
            priorityStyles = elementPriority(series);

        for (var attr in currentOrigProp) {
            if (currentOrigProp.hasOwnProperty(attr) && !arrayIncludes(priorityStyles, attr)) {
                var val = currentOrigProp[attr],
                    actualVal = function() {
                        return multipleStyles ? val[count++] : val;
                    },
                    count = 0;
                if ($Browser.SAFARI || $Browser.IE) {
                    series.style(attr, actualVal).attr(attr, actualVal);
                } else {
                    series.style(attr, actualVal);
                }
            }
        }
    }

    function getKey(series, index, selectionType, paneID) {
        /*
         * This function returns a unique key that is generated from the localName, class, id of the series element and index.
         * The key will be used to store the style values of the element on legend highlighting. The stored values will be applied to the element on unhighlighting.
         */
        var node = series.node();
        var localName = node.localName,
            className = node.getAttribute('class'),
            idName = node.getAttribute('id');
        var KEY = `${localName}_${index}`;
        if (className != null) {
            KEY += `_${className}`;
        }
        if (idName != null) {
            KEY += `_${idName}`;
        }
        if (selectionType !== null) {
            KEY += `_${selectionType}`;
        }

        if (paneID !== null) {
            KEY += `_${paneID}`;
        }
        return KEY;
    }

    function setSeriesVisible(series, visible) {
        series.style('visibility', visible ? 'visible' : 'hidden');
    }

    function defaultEffect(chartName) {
        if (L_H.hasPatternPallete(chartObj)) {
            return /(area|arearange)/.test(chartName) ? 'moody' : 'strokeEnlarge';
        }

        if (/(bar|bullet|gantt|area|arearange|pie|funnel|pyramid|web|sunburst|waterfall|boxplot|sankey|barrange|dial)/.test(chartName)) {
            return 'hatchfill';
        } else if (chartName === 'heatmap' || chartName === 'geoheatmap') {
            return 'tinyHatchfill';
        } else if (/(treemap|wordcloud)/.test(chartName)) {
            return 'blackyBorder';
        } else {
            return 'strokeEnlarge';
        }
    }

    function invalidEffect(effect, chartName) {
        return /(line|scatter)/.test(chartName) && /(((\D)+fill$)|illuminate|moody|flip)/.test(effect);
    }

    function highlightSVGSeries(legdata, index, filterIndex, isSelection) {
        legendHighlight.selSeriesObj = [];
        index = splat(index);
        index.forEach(function(ii) {
            var selected = getSelectedSeries(ii, filterIndex, isSelection);//the parameter is passed to add ellipse component to funel selection effect
            legendHighlight.selSeriesObj.push(selected);

            var selectedSeries = selected.selectedSeries;
            var charttypes = selected.charttypes;
            var parentSeries = selected.parentSeries;
            var paneIDs = selected.paneIDs;
            var length = selectedSeries.length,
                prop;

            for (var i = 0; i < length; i++) {
                var selecteffect = legendHighlight.getLegendHighlightEffect(ii, isSelection ? 'selection' : 'selectedSeries', undefined, paneIDs[i]);
                var cType = charttypenames.get(charttypes[i]),
                    baseColor = legendHighlight.baseClr || getCurrectColor(ii, filterIndex);

                if (chartObj.dataObject.isFlowChartCategory) {
                    // eslint-disable-next-line no-loop-func
                    selectedSeries[i].each(function(p) {
                        // every sankey point has different colors. so getting the color for individuals points
                        baseColor = chartObj.renderer[0].getColor(p.point);
                        // baseColor = p.point.arcColor;
                        if(isSelection){
                            prop = {
                                'fill' :chartObj.systemConf.chart.plot.eventHandler.click.behaviour.unselected
                            }
                        }
                        else{
                            prop = legendHighlight.getEffect(selecteffect, cType, baseColor, null, ii, null, p);
                        }
                        applyProp(d3_select(this), parentSeries[i], prop, p.point.itemIndex, isSelection ? 'selection' : 'select');
                    });
                } else {
                    // TODO below line
                    // baseColor = legendHighlight.baseClr || getCurrectColor(ii, filterIndex);
                    // const element = parentSeries[i]? parentSeries[i].node() : null;
                    const additionalStyles = getHoverGradients(chartObj.seriesdata[ii], ii, chartObj);
                    if(isSelection){//to apply colour as selection pattern 
                        switch(cType){
                            case 'treemap':
                            case 'sunburst':
                            case 'pie':
                                prop = {
                                    'fill' : chartObj.systemConf.chart.plot.eventHandler.click.behaviour.unselected,
                                    // 'stroke' : chartObj.systemConf.chart.plot.eventHandler.click.behaviour.unselected
                                };
                                break;
                            case 'scatter':
                            case 'bubble':
                            case 'bar':
                            case 'funnel':
                                prop = {
                                    'fill' : chartObj.systemConf.chart.plot.eventHandler.click.behaviour.unselected,
                                    'stroke' : chartObj.systemConf.chart.plot.eventHandler.click.behaviour.unselected
                                };
                                break;
                            case 'line':
                                prop = {
                                    'stroke' : chartObj.systemConf.chart.plot.eventHandler.click.behaviour.unselected
                                };
                                break;
                            case 'area':
                                prop = {
                                    'fill' : chartObj.systemConf.chart.plot.eventHandler.click.behaviour.unselected,
                                };
                                break;
                        }
                    }
                    else{
                        prop = legendHighlight.getEffect(selecteffect, cType, baseColor, additionalStyles, ii, null, null, null, selectedSeries[i]);
                    }
                    applyProp(selectedSeries[i], parentSeries[i], prop, ii, (isSelection)? 'selection' : 'select', paneIDs[i]);
                }
            }
        });
    }

    function highlightCanvasSeries(chartObj, legdata, index, filterIndex, isSelection) {
        var context =
                chartObj.highlightContext ||
                (chartObj.highlightContext = plot.canvasRenderer.updateCanvasContext(
                    chartObj,
                    (isSelection) ? 'selection-canvas' : 'highlight-canvas',
                    null,
                    chartObj.canvasScrollGroup
                )),
            scrollPosition = [0, 0];

        if (chartObj.canvasScrollGroup) {
            d3_select(context.canvas).styles({
                left: '0px',
                top: '0px'
            });
            scrollPosition = [-chartObj.scrolledX, -chartObj.scrolledY]; // ZC4238
        }
        index = splat(index);

        var canvasSeriesHighlight = function(index, seriesdata, rendererIndex, layerIndex, seriesIndex, paneID) {
            let chartTypeObj = getChartTypeByIndex(chartObj, paneID, index),
                chartName = chartTypeObj.name,
                cType = chartTypeObj.value,
                isAxisCategory = isCategoryOf(chartObj, cType, 'axis'),
                isPolarAxisCategory = isCategoryOf(chartObj, cType, 'polarAxis'),
                isNonAxisMultiSeriesCategory = isCategoryOf(chartObj, cType, 'nonAxisMultiSeries');

            var effect = legendHighlight.getLegendHighlightEffect(index, (isSelection) ? 'selection' : 'selectedSeries', undefined, paneID),
                unHighlightEffect = legendHighlight.getLegendHighlightEffect(index, 'unselectedSeries', undefined, paneID);
            seriesIndex = pick(seriesIndex, isAxisCategory || isNonAxisMultiSeriesCategory || chartObj.isGeoMap ? index : 0);

            if (effect === NONE && !UNHIGHLIGHT_EFFECT_REGEX.test(unHighlightEffect)) {
                return;
            }

            if(isSelection && chartName == 'sankey' && chartObj.cache.sankey && chartObj.cache.sankey.legend[index].disabled){
                return;
            }

            var renderer = plot.canvasRenderer.createCanvasRenderer(chartName),
                currentSeries = seriesdata[seriesIndex],
                pattern = null,
                highlightSeriesindex =
                    (isNonAxisMultiSeriesCategory && !isPolarAxisCategory) || chartObj.dataObject.isFlowChartCategory ? 0 : index,
                fillOpacity = chartObj.systemConf.chart.plot.plotoptions[chartName].fillOpacity;

            let gradient = null;

            if (encode === 'z' && !(chartcategory.zaxis.indexOf(cType) > -1 && !isHeatMapChartType(cType))) {
                // Size legend highlight: non-zaxis charts need not to drawn
                return;
            }
            rendererIndex = pick(rendererIndex, highlightSeriesindex);

            var rp = chartObj.renderer[rendererIndex].commonRendererProp().params,
                legendIndex = isFacet(chartObj) ? rp.legendIndex : index,
                baseColor = getUserFillColor(index, chartName) || getCurrectColor(legendIndex, filterIndex) || legendHighlight.baseClr;

            renderer_helpers.setRendererProp(
                chartObj,
                renderer,
                seriesdata,
                highlightSeriesindex,
                highlightSeriesindex,
                {
                    labelQueue: [],
                    labelArray: {}
                },
                rendererIndex,
                DataProcessor.helpers.getChartType(
                    (isFacet(chartObj) ? currentSeries.node.type : currentSeries.type) || chartObj.userdata.seriesdata.type,
                    NUMERIC_DATATYPE
                ),
                layerIndex
            );

            if (hasColorScale && (isAreaChartType(cType) || isScatterChartType(cType))) {
                pattern = legendHighlight.getEffect(effect, cType);
            } 
            else if(isSelection){// To apply selection colour as pattern     
                switch(cType){
                    case 36:
                        pattern = {//sankey
                            'fill' : chartObj.systemConf.chart.plot.eventHandler.click.behaviour.unselected,
                            'fill-opacity': chartObj.systemConf.chart.plot.plotoptions.sankey.link['fillOpacity']
                        };
                        break;
                    default:
                        pattern = chartObj.systemConf.chart.plot.eventHandler.click.behaviour.unselected;
                        break;
                }
            }
            else {
                switch (chartName) {
                    case 'bar':
                    case 'bullet':
                    case 'area':
                    case 'arearange':
                    case 'heatmap':
                    case 'treemap':
                    case 'web': {
                        var prop = patternThemes[effect] ? getPattern(effect) : null;
                        if (prop) {
                            gradient = getHoverGradients(currentSeries, index, chartObj, chartObj.plot.renderer.mode).gradient;
                            gradient ? (prop.bgClr = prop.bgClr.split(',').map(d => 'transparent').join(',')) : null;
                            pattern = getCanvasPattern(prop, baseColor, fillOpacity, chartObj);
                            gradient ? (pattern =  { pattern, gradient }) : null;
                        }
                        break;
                    }
                    case 'sankey':
                    case 'wordcloud': {
                        var prop = patternThemes[effect] ? getPattern(effect) : null;
                        if (prop) {
                            pattern = { fill: getCanvasPattern(prop, baseColor, fillOpacity, chartObj) };
                        } else {
                            pattern = legendHighlight.getEffect(effect, cType);
                        }
                        break;
                    }
                    case 'scatter':
                    case 'geoscatter':
                    case 'bubble':
                    case 'geobubble':
                    case 'line': {
                        // TODO CHECK
                        // getEffect expect a chart name but we were giving as number
                        pattern = legendHighlight.getEffect(effect, chartName);
                    }
                }
            }
            var dataForCanvas = isAxisCategory || isFacet(chartObj) ? seriesdata : currentSeries;

            let { bound } = rp,
                x = scrollPosition[0] + (isAxisCategory ? 0 : bound.x + (isPolarAxisCategory ? bound.polar.x : 0)),
                y = scrollPosition[1] + (isAxisCategory ? 0 : bound.y + (isPolarAxisCategory ? bound.polar.y : 0));

            context.translate(x, y);
            plot.canvasRenderer.drawCanvas(
                chartObj,
                dataForCanvas,
                renderer,
                context,
                pattern,
                filterIndex,
                encode,
                (chartTypeObj.name === 'sankey' && chartObj.systemConf.legend.ranges) ? filterIndex: chartObj.dataObject.isFlowChartCategory ? index : seriesIndex,
                isSelection
            );
            context.translate(-x, -y);
        };

        index.forEach(function(ii) {
            if (isFacet(chartObj)) {
                var gridList = getGridList(chartObj.dataObject),
                    count = 0;

                gridList.forEach(grid => {
                    let { layer, row, column } = grid,
                        paneID = `${row}_${column}`;

                    grid.data.forEach((_d, _i) => {
                        let chartType = getChartTypeByIndex(chartObj, paneID, _i).value,
                            isAxisCategory = isCategoryOf(chartObj, chartType, 'axis'),
                            isPolarAxisCategory = isCategoryOf(chartObj, chartType, 'polarAxis');

                        if (isAxisCategory || isPolarAxisCategory) {
                            if (
                                (!chartObj.userdata.facets.color.length || layer === order) &&
                                chartObj.rendererConf[count].legendIndex() === ii
                            ) {
                                canvasSeriesHighlight(_i, grid.data, count, _d.layer, undefined, paneID);
                            }
                        } else {
                            if (
                                (!chartObj.userdata.facets.color.length || layer === order) &&
                                chartObj.rendererConf[count].legendIndex() === ii
                            ) {
                                let _rendererIndex = count - _i;
                                canvasSeriesHighlight(
                                    ii,
                                    grid.data,
                                    _rendererIndex,
                                    _d.layer,
                                    chartObj.rendererConf[count].datasetIndex(),
                                    paneID
                                );
                            }
                        }

                        count++;
                    });
                });
            } else {
                canvasSeriesHighlight(ii, chartObj.seriesdata);
            }
        });
    }

    /**
     *
     * @param chartObj
     * @param legdata
     * @param index
     * @param filterIndex
     */
    function highlightGeoMap(chartObj, legdata, index, filterIndex) {
        if (legendHighlight.previousFilterIndex !== filterIndex) {
            var mapEvents = chartObj.eventHandler.mapEvents,
                points = mapEvents.getPointsByIndex(index, filterIndex),
                features = chartObj.dataObject.getMapFeaturesByDataPoint(points, 'array'),
                highlightStyle = mapEvents.getHighlightEffect(index, null, 'selectedSeries', 'legend');

            mapEvents.highlightPoints(features, 'feature', highlightStyle);
            legendHighlight.previousFilterIndex = filterIndex;
        }
    }

    function unHighlightSVGSeries(legdata, index, filterIndex, isSelection) {
        index = splat(index);
        var unselSeriesObj = getUnselectedSeries(index, filterIndex);
        var unselectedSeries = unselSeriesObj.unselectedSeries;
        var indices = unselSeriesObj.indices;
        var charttypes = unselSeriesObj.charttypes;
        var paneIDs = unselSeriesObj.paneIDs;
        var length = unselectedSeries.length;
        for (var i = 0; i < length; i++) {
            var cType = charttypenames.get(charttypes[i]);
            var unselecteffect = legendHighlight.getLegendHighlightEffect(indices[i], isSelection ? 'selection' : 'unselectedSeries', undefined, paneIDs[i]);

            if (chartObj.dataObject.isFlowChartCategory) {
                // eslint-disable-next-line no-loop-func
                unselectedSeries[i].each(function(p) {
                    // every sankey point has different colors. so getting the color for individuals points

                    prop = legendHighlight.getEffect(unselecteffect, cType, NULL, NULL, indices[i]);
                    applyProp(d3_select(this), NULL, prop, p.point.itemIndex, 'unselect');
                });
            } else {
                var prop = legendHighlight.getEffect(unselecteffect, cType, NULL, NULL, indices[i]);
                applyProp(unselectedSeries[i], NULL, prop, indices[i], isSelection ? 'selection' : 'unselect', paneIDs[i]);
            }
        }
    }

    function unHighlightCanvasSeries(legdata, index, filterIndex) {
        var ele = d3_select(chartObj.context.canvas),
            highlightEffect = legendHighlight.getLegendHighlightEffect(0, 'unselectedSeries'),
            prop = legendHighlight.getEffect(highlightEffect, NULL, NULL, NULL, 0);
        applyProp(ele, NULL, prop, 0, 'unselect');
    }

    function resetSelectedCanvasSeries(legdata, index, filterIndex) {
        chartObj.container.select('.highlight-canvas').style('visibility', 'hidden');
    }

    function resetUnselectedCanvasSeries(legdata, index, filterIndex) {
        var ele = d3_select(chartObj.context.canvas),
            highlightEffect = legendHighlight.getLegendHighlightEffect(0, 'unselectedSeries'),
            prop = legendHighlight.getEffect(highlightEffect, NULL, NULL, NULL, 0);
        resetProp(ele, prop, 0, 'unselect');
    }

    function resetGeoMapSeries() {
        legendHighlight.previousFilterIndex = null;
        chartObj.eventHandler.mapEvents.clearHighlightedPoints();
    }

    function highlightNonSVGSeries(legdata, index) {
        var selecteffect = legendHighlight.getLegendHighlightEffect(index, 'selectedSeries');
        if (selecteffect === NONE) {
            return;
        }

        legendHighlight.baseClr = getColorByIndex(chartObj, index, null, order);

        var cType = getChartTypeByIndex(chartObj, '0_0', index);
        var cName = cType.name;
        var cTypeNumeric = cType.value;

        var selectedSeriesObj = getSelectedSeries(index);

        if (selectedSeriesObj.selectedSeries[0].length > 0) {
            setSeriesVisible(selectedSeriesObj.selectedSeries[0], true);
            return;
        }

        if (!chartObj.chartRendered) {
            return;
        }

        /*
         * Disable animation during the chartObj drwaing proccess and disable the datalabels also.
         */

        chartObj.systemConf.chart.plot.animation.enabled = false;
        chartObj.systemConf.chart.plot.plotoptions[cName].animation.enabled = false;
        if (chartObj.seriesdata[index].plotoptions && chartObj.seriesdata[index].plotoptions.animation) {
            chartObj.seriesdata[index].plotoptions.animation.enabled = false;
        }

        var renderer = plot.svgRenderer.createSVGRenderer(cName, chartObj);
        var currOrdereddata = chartObj.systemConf.seriesdata.chartdata[index];
        var barDatasetIndex = 0;
        if (cTypeNumeric === charttype.bar || cTypeNumeric === charttype.stkdbar) {
            for (var i = 0; i < index; i++) {
                cType = getChartTypeByIndex(chartObj, '0_0', i);
                cTypeNumeric = cType.value;
                if ((chartObj.seriesdata[i].disabled == undefined || !chartObj.seriesdata[i].disabled) && isBarChartType(cTypeNumeric)) {
                    barDatasetIndex = barDatasetIndex + chartObj.seriesdata[i].data.length;
                }
            }
        }
        renderer_helpers.setRendererProp(chartObj, renderer, currOrdereddata, index, index, {
            labelQueue: [],
            barDatasetIndex: barDatasetIndex
        });
        plot.svgRenderer.drawSVG(chartObj, currOrdereddata, renderer);

        highlightSVGSeries(legdata, index);
        return true;
    }

    function hasMultipleStyles(seriesIndex) {
        var cType = getChartType(null, NUMERIC_DATATYPE, seriesIndex, userdata);

        if (cType === charttype.sankey) {
            return false;
        }

        return (
            chartcategory.hierarchy.indexOf(cType) != -1 ||
            hasColorScale ||
            isBubbleChartType(cType) ||
            (chartObj.rendererConf && chartObj.rendererConf[chartObj.isAxisCategory ? seriesIndex : 0].thresholdType)
        );
    }

    function doArrangeSelection(paneID = '0_0') {
        let cType_obj = getChartTypeByIndex(chartObj, paneID, 0),
            dataObject = chartObj.dataObject;

        if (dataObject.isFlowChartCategory) {
            return true;
        } else if (isCategoryOf(chartObj, cType_obj.value, 'hierarchy')) {
            var styles = legendHighlight.getEffect(legendHighlight.getLegendHighlightEffect(0, 'selectedSeries'), cType_obj.name);
            var keys = Object.keys(styles);

            return (
                keys.length > 0 &&
                keys.some(function(key) {
                    return key !== 'fill';
                })
            );
        } else if (
            isCategoryOf(chartObj, cType_obj.value, 'axis') &&
            (dataObject.noofAreaEnabledSeries || dataObject.noofAreaRangeEnabledSeries)
        ) {
            return true;
        }
    }

    function hasSubScatterRenderer(filterValue) {
        return $Browser.IE || $Browser.SAFARI || defined(filterValue);
    }

    function getCurrectColor(index, filterIndex) {
        // let _custom = getUserFillColor(index, chartName);
        //
        // if (_custom) {
        //     return _custom;
        // }
        if (defined(filterIndex) && encode === 'clr') {
            var userRanges = chartObj.systemConf.legend.colorBand.ranges;
            if (userRanges && userRanges.length) {
                return getColorByIndex(chartObj, filterIndex, null, order);
            } else {
                if(chartObj.axes.clr.scale){
                    return chartObj.axes.clr.scale((filterIndex[0] + filterIndex[1]) / 2);
                }else{
                    return getColorByIndex(chartObj, index, null, order);
                }
                
            }
        } else if (encode === 'z') {
            return (d, i) => getRawColor(chartObj, i, d && (d.data || d), index);
        } else {
            return getColorByIndex(chartObj, index, null, order);
        }
    }

    function getGeoHighlightPattern(chartObj, index, filterIndex, baseColor) {
        var effect = legendHighlight.getLegendHighlightEffect(index, 'selectedSeries'),
            chartName = getChartTypeByIndex(chartObj, '0_0', index).name,
            fillOpacity = chartObj.systemConf.chart.plot.plotoptions[chartName].fillOpacity,
            prop = getPattern(effect);

        if (prop) {
            return getCanvasPattern(prop, baseColor || legendHighlight.baseClr, fillOpacity);
        }
    }

    function getUserFillColor(index, chartName) {
        if (isFacet(chartObj)) {
            return;
        }

        if (arrayIncludes([AREA_CHARTNAME, AREARANGE_CHARTNAME, WEB_CHARTNAME], chartName)) {
            let _data = chartObj.seriesdata[index],
                commonPlotOptions = chartObj.systemConf.chart.plot.plotoptions[chartName];
            return isObjectPropDefined(_data, 'plotoptions.fillColor', true) || commonPlotOptions.fillColor;
        }
    }

    return legendHighlight;
};

function getCanvasPattern(prop, baseClr, fillOpacity, chartObj) {
    var isSelectionContainer = !(chartObj.container.select('.selection-canvas').empty());
    prop.bgClr_base = isSelectionContainer ? 'rgba(0,0,0,0)' : rgba_canvas(baseClr, fillOpacity);
    prop.stroke = 'transparent,white,white'; //need to check
    return effects.get(effects.canvas_patterns, 'canvasPatternFill', prop, { chartObj: chartObj });
}

function noColorColumn(chartObj) {
    return (
        !isFacet(chartObj) ||
        (isFacet(chartObj) &&
            arrayIncludes([PIE_CHARTNAME, FUNNEL_CHARTNAME, PYRAMID_CHARTNAME], chartObj.chartTypes[0].name) &&
            !chartObj.userdata.facets.color.length)
    );
}

Registry.setComponent('LegendHighlight', legendHighlight);
