// @zohocharts/zc/src/legend/legend-events.js
//$Id$
//legend-events.js

import { legend, legend_helpers } from '../core/main-zc-init';
import { hasAssociatedLegendData, getLegendContainer } from './legend-helper';
import { getValByPriority, validateNullVal, invokeFunction, isEmptyObject, defined, splat, arrayIncludes } from '../utils/zcutils';
import { POINTER, NULL } from '../core/zc-constants';
import { applyFont } from '../utils/zcsvgutils';
import { d3_select, d3_event, d3_set } from '../core/d3-constants';
import { colorscale_helpers } from '../axis/colorscale-helpers';
import { attachevents } from '../events/eventActions';
import { $ZCG } from '../core/zcglobal';
import { docTapFunction } from '../events/document-events';
import { getLegendShowAsText } from './legend-utils';
import { $Browser } from '../core/browser-detect';
import { DataProcessor } from '../core/data-processor';
import { getHoverOptions, getDisabledOptions } from './events/legend-event-helpers';
import { SANKEY_CHARTNAME } from '../core/chartname-constants';
import { getEmptySankeyNode, getSankeyDependentPoint } from './events/data-selection/sankey/sankey-dependent-point';
import { isFacet } from '../axis/hierarchy-axis-utils';
import { getGridList } from '../core/data-processor/pivot-layout-helpers';
import { customGestureList } from '../events/custom-gesture';
import { Registry } from '../core/zc-registry';
import { SERIES_LEGEND_CLASSNAME } from './legend-constants';
import { isWithinColorRange } from '../utils/colorutils';

var getLegendData = legend_helpers.getLegendData,
    getLegendDataLength = legend_helpers.getLegendDataLength,
    isHiddenItem = legend_helpers.isHiddenItem,
    isDisabled = legend_helpers.isDisabled,
    isItemDisabled = legend_helpers.isItemDisabled,
    isAllEnabled = legend_helpers.isAllEnabled,
    getClr = legend_helpers.getClr,
    updateData = legend_helpers.updateData;

legend.events = function(chartObj, element, states, disabledOpacity, order = 0) {
    var legendComponent = chartObj.legend.component,
        legendOptions = chartObj.systemConf.legend,
        hoverOptions = getHoverOptions(legendOptions, 'clr'),
        disabledOptions = getDisabledOptions(legendOptions, 'clr'),
        colorBoxOptions = legendOptions.colorBox,
        priorityArray = [legendOptions, chartObj.systemConf.canvas, $ZCG],
        priorityArrayOnhover = [hoverOptions, legendOptions, chartObj.systemConf.canvas, $ZCG],
        fontColor = getValByPriority(priorityArray, 'fontColor'),
        legendFilter,
        selectAllFilter,
        legData,
        legData_length,
        chartName = chartObj.chartTypes[0].name,
        isSankeyChart = chartName === SANKEY_CHARTNAME,
        associatedData = [];

    /*
     * legend event variables
     */
    var events_legend = chartObj.systemConf.legend.events;
    var onClickFn, onDblClickFn, onMouseOverFn, onMouseOutFn, onTapFn, onTapHoldFn, onTapHoldEndFn, cursor, eventDisabled;
    if (events_legend != null) {
        eventDisabled = validateNullVal(events_legend.enabled, true) == false;
        onClickFn = events_legend.click;
        onDblClickFn = events_legend.doubleclick;
        onMouseOverFn = events_legend.mouseover;
        onMouseOutFn = events_legend.mouseout;
        cursor = events_legend.cursor;
        onTapFn = events_legend.tap;
        onTapHoldFn = events_legend.taphold;
        onTapHoldEndFn = events_legend.tapholdend;
    }

    if (!eventDisabled) {
        element && element.style('cursor', cursor || POINTER);

        var eventAttacherOptions = {
            chartObj: chartObj
        };
        var mousemove = function(d, i) {
                $ZCG.clickedTime = null;
                if (onMouseOverFn == null) {
                    defaultOnMouseOverFunc(this, d, i);
                } else {
                    var userFunction = chartObj.zcEventType ? onTapHoldFn : onMouseOverFn;
                    invokeFunction(userFunction, d3_event(), d, i, chartObj);
                }
            },
            mouseout = function(d, i) {
                if (onMouseOutFn == null) {
                    defaultOnMouseOutFunc(this, d, i);
                } else {
                    var userFunction = chartObj.zcEventType ? onTapHoldEndFn : onMouseOutFn;
                    invokeFunction(userFunction, d3_event(), d, i, chartObj);
                }
            },
            click = function(d, i) {
                var isTouchEvent = d3_event() && d3_event().touches;
                if (isTouchEvent) {
                    docTapFunction();
                }
                if (onClickFn == null) {
                    defaultOnClickFunc(this, d, i);
                } else {
                    var userFunction = chartObj.zcEventType ? onTapFn : onClickFn;
                    invokeFunction(userFunction, d3_event(), d, i, chartObj);
                }
            },
            dblclick = function(d, i) {
                if (onDblClickFn == null) {
                    var dummy = true;
                    //default double click function
                } else {
                    invokeFunction(onDblClickFn, d3_event(), d, i, chartObj);
                }
            };

        let tempEventName = $Browser.PHANTOMJS
            ? {
                  move: 'mousemove',
                  out: 'mouseout'
              }
            : {
                  move: 'mouseenter',
                  out: 'mouseleave'
              };

        var mouseHandlers = {
                // TODO we should consider mouseenter and mouseleave for better performance
                // we are currently not using it because phantomjs does not support it and legend event automation does not work
                // mouseenter: mousemove,
                // mouseleave: mouseout,
                [tempEventName.move]: mousemove,
                [tempEventName.out]: mouseout,
                click: click,
                dblclick: dblclick
            },
            touchHandlers = {
                taphold: mousemove,
                tapholdend: mouseout,
                tap: click,
                dbltap: dblclick
            };
        // if(element){
            // attachevents(element, mouseHandlers, touchHandlers, eventAttacherOptions);
        // }
        

        var callCustomHandler = (eventType, targetEvent, d, i, element) => {
            if (events_legend && events_legend[eventType]) {
                invokeFunction(events_legend[eventType], targetEvent, d, i, chartObj, element);
            }
        };

        if (customGestureList) {
            let gestureKeys = Object.keys(customGestureList),
                CustomGesture = Registry.getComponent('CustomGesture');
            gestureKeys.map(d => {
                CustomGesture.call(this, d, element, 'legend', callCustomHandler);
            });
        }
    }

    function defaultOnClickFunc(ele, d, i) {
        legendComponent.toggleItemNUpdateChart(ele, d, i, legendOptions);
    }

    function defaultOnMouseOverFunc(ele, d, i) {
        legendComponent.highlightSeries(ele, d, i, legendOptions, chartObj, true);
        legendComponent.highlightItem(ele, d, i, legendOptions, true);
    }

    function defaultOnMouseOutFunc(ele, d, i) {
        legendComponent.highlightSeries(ele, d, i, legendOptions, chartObj);
        legendComponent.highlightItem(ele, d, i, legendOptions);
    }

    legendComponent.filterSeriesNUpdate = function(d, i, chartObj) {
        var indices = splat(i),
            legendFilterAllowed = allowLegendFilter(d, i);

        initLegendData();

        if (isAlldisabled(indices, legendFilterAllowed) || !legendFilterAllowed) {
            return;
        }

        indices.forEach(i => {
            if (hasAssociatedLegendData(chartObj)) {
                var gridList = getGridList(chartObj.dataObject),
                    count = -1;

                gridList.forEach((grid, _gridIndex) => {
                    let { layer } = grid;

                    grid.data.forEach((_d, _i) => {
                        count++;
                        if (
                            (!chartObj.userdata.facets.color.length || layer === order) &&
                            chartObj.rendererConf[count].legendIndex() === i
                        ) {
                            _d.node.filtered = !_d.node.filtered;
                        }

                        if (layer === order) {
                            legData[chartObj.rendererConf[count].legendIndex()].filtered = _d.node.filtered;
                        }
                    });
                });
                //TODO: check bubblepie
            } else {
                var d = legData[i];
                d.disabled = !d.disabled;

                if ((chartObj.dataObject.hierarchical) && !colorscale_helpers.isQuantileColorScale(chartObj)) {
                    chartObj.systemConf.seriesdata.chartdata[i].disabled = d.disabled;
                }

                updateData(d, i, chartObj);
            }
        });

        var distinctAxes = getDistinctAxes(indices);
        distinctAxes.forEach(i => {
            var dt = legData[i];
            chartObj.addRemoveSeries(dt, i, dt.yaxiscolumnorder, !states.allSelected && !states.isFirstClickAfterDeselection);
        });
        updateStates();
    };

    legendComponent.toggleItemNUpdateChart = function(ele, d, i, legendOptions) {
        var indices = splat(i);
        legendFilter = getLegendContainer(chartObj, `clr-${order}`).selectAll('div.legendfilterbox, div.zc-legend-legendfilterbox');

        if (isLegendFilterEnabled()) {
            var highlightObj = getLegendHighlightComp();
            highlightObj.resetSelection();
            legendComponent.filterSeriesNUpdate(d, indices, chartObj);

            // Reset the hoverstyle for disabled series; strictApply=true
            var disabledIndices = indices.filter(i => isItemDisabled(legData[i]));
            legendComponent.highlightItem(ele, d, disabledIndices, legendOptions, false, true);

            // update all the connected labels
            if (isSankeyChart) {
                let indices = legData.map((l, i) => l.itemIndex? l.itemIndex : i);
                legendComponent.updateLabel(indices);
            } else {
                // Paint legend labels
                legendComponent.updateLabel(indices);
            }

            // update the colorBox
            legend.setColorBoxBackground(legendFilter, chartObj, states, order);

            // Mouse actions for non-touch devices
            if (d3_event() && !d3_event().touches) {
                if (chartObj.plot.renderer.mode === 'canvas') {
                    legendComponent.highlightSeries(ele, d, indices, legendOptions, chartObj); //reset canvas' unhighlighted effects
                }

                var mouseAction = isItemDisabled(legData[indices[0]]) ? defaultOnMouseOutFunc : defaultOnMouseOverFunc;
                mouseAction(ele, d, indices);
            }
        }

        updateSelectAllBox();
    };

    legendComponent.highlightItem = function(ele, d, i, legendOptions, onLegHover, strictApply) {
        var indices = splat(i);
        initLegendData();

        indices.forEach(index => {
            var d = legData[index],
                ele = getLegendItem(index).node();

            _highlightItem(ele, d, index, legendOptions, onLegHover, strictApply);
        });
    };

    function _highlightItem(ele, d, i, legendOptions, onLegHover, strictApply) {
        if (strictApply || !isDisabled(d, states.allSelected, states.isFirstClickAfterDeselection)) {
            var hoverOptions = getHoverOptions(legendOptions, 'clr');

            if (!isEmptyObject(hoverOptions)) {
                var filteredEle = d3_select(ele),
                    legendLabel = filteredEle.selectAll('.legendlabel, .zc-legend-legendlabel'),
                    backgroundColor = (onLegHover ? hoverOptions : legendOptions).backgroundColor;
                filteredEle.style('background-color', backgroundColor);
                if (!onLegHover && defined(legendOptions.showAs)) {
                    legendLabel.html(getLegendShowAsText(d, i, legendOptions, chartObj));
                }
                applyFont(
                    filteredEle.selectAll(onLegHover ? '.legendlabel, .legendlabel *, .zc-legend-legendlabel, .zc-legend-legendlabel *' : '.legendlabel, .zc-legend-legendlabel'),
                    onLegHover ? priorityArrayOnhover : priorityArray,
                    true
                );
            }
        }
    }

    legendComponent.highlightSeries = function(ele, d, i, legendOptions, chartObj, onLegHover, isSelection) {
        var highlightObj = getLegendHighlightComp();
        chartObj.highlightContext = null;
        // const legendStates = chartObj.legend.states; // identifies a series is highlighted or not 
        // legendStates.seriesHighlighted = true;
        legendComponent.handlePreviousHighlight.apply(legendComponent, arguments);

        if (colorscale_helpers.isQuantileColorScale(chartObj)) {
            var filterIndex = i;
            if (isFacet(chartObj)) {
                if (onLegHover) {
                    if (!isItemDisabled(d)) {
                        highlightObj.highlightSelectedSeries(legendOptions, 0, filterIndex);
                        highlightObj.highlightUnselectedSeries(legendOptions, 0, filterIndex);
                    }
                } else {
                    highlightObj.reset(legendOptions, 0, filterIndex);
                }
            } else {
                if (onLegHover) {
                    if (!isItemDisabled(d)) {
                        chartObj.seriesdata.forEach(function(d, j) {
                            highlightObj.highlightSelectedSeries(legendOptions, j, filterIndex);
                            highlightObj.highlightUnselectedSeries(legendOptions, j, filterIndex);
                        });
                    }
                }
                else if(isSelection){
                    highlightObj.highlightSelectedSeries(legendOptions, i, null, isSelection);
                }
                else {
                    chartObj.seriesdata.forEach(function(d, j) {
                            highlightObj.reset(legendOptions, j, filterIndex, chartObj.seriesdata.length - 1 === j);
                    });
                }
            }
        } else {
            if (onLegHover) {
                if (!isItemDisabled(d)) {
                    highlightObj.highlightSelectedSeries(legendOptions, i);
                    highlightObj.highlightUnselectedSeries(legendOptions, i);
                }
            } 
            else if(isSelection){
                highlightObj.highlightSelectedSeries(legendOptions, i, null, isSelection);
            }
            else {
                highlightObj.reset(legendOptions, i);
            }
        }
        // legendStates.seriesHighlighted = false;
    };

    legendComponent.updateLabel = function(i) {
        var indices = splat(i);
        indices.forEach(index => _updateLabel(index));
    };

    legendComponent.handlePreviousHighlight = function() {
        var highlightObj = getLegendHighlightComp(),
            prev = highlightObj.prevData,
            toggleIndex = 5,
            onLegHover = arguments[toggleIndex];

        // ZC1758: Reset previous highlight effects
        if (onLegHover && prev) {
            prev[toggleIndex] = false;
            highlightObj.prevData = NULL;
            legendComponent.highlightSeries.apply(legendComponent, prev);
        }
        highlightObj.prevData = onLegHover ? arguments : NULL;
    };

    function _updateLabel(i) {
        var d = legData[i],
            legendItem = getLegendItem(i),
            labelFontColor = getClr(isItemDisabled(d), fontColor, disabledOptions),
            labelDecoration = isItemDisabled(d) && disabledOptions.strikeout ? 'line-through' : null;

        legendItem.selectAll('.legendlabel, .legendlabel *, .zc-legend-legendlabel, .zc-legend-legendlabel *').styles({
            color: labelFontColor,
            'text-decoration': labelDecoration
        });

        // for custom legend, while enabling legend items, get the source from the user.
        // because user might set some different colors and can have different hierarchy
        if (!isItemDisabled(d) && legendOptions.showAs) {
            legendItem.selectAll('.legendlabel, .zc-legend-legendlabel').html(getLegendShowAsText(d, i, legendOptions, chartObj));
        }
    }

    function getLegendItem(i) {
        var ele = chartObj.legends[order].element.selectAll(`li.${SERIES_LEGEND_CLASSNAME}`);
        return d3_select(ele.nodes()[i]);
    }

    function initLegendData() {
        if (legData) {
            return;
        }

        //If legend is disabled, legData will not be available
        legData = getLegendData(chartObj, order);
        legData_length = getLegendDataLength(chartObj, legData);

        if (hasAssociatedLegendData(chartObj)) {
            var gridList = getGridList(chartObj.dataObject),
                count = -1;

            gridList.forEach((grid, _gridIndex) => {
                let { layer } = grid;

                grid.data.forEach((_d, _i) => {
                    count++;
                    if (layer === order) {
                        let legendIndex = chartObj.rendererConf[count].legendIndex();
                        associatedData[legendIndex] = associatedData[legendIndex] || [];
                        associatedData[legendIndex].push(_d);
                    }
                });
            });
        }
    }

    function updateStates() {
        if (!states.allSelected && !states.isFirstClickAfterDeselection) {
            states.isFirstClickAfterDeselection = true;
        }
    }

    function isAlldisabled(indices, legendFilterAllowed) {
        let sankeyDependentPoints,
            sankeyEmptyNodes,
            currentLegendPoint = legData[indices[0]],
            enabling = isItemDisabled(currentLegendPoint) === true;

        if (isSankeyChart) {
            //TODO:= Work on sankey hierarchy legend filtering
            sankeyDependentPoints = getSankeyDependentPoint(currentLegendPoint, chartObj, states);
            sankeyEmptyNodes = getEmptySankeyNode(currentLegendPoint, chartObj).filter(p => !arrayIncludes(sankeyDependentPoints, p));
        }

        var disabledEle = legData.filter(function(d, j) {
                if (!states.allSelected && !states.isFirstClickAfterDeselection) {
                    // sankey dependent points check
                    if (associatedData && associatedData.length) {
                        associatedData[j].forEach(_d => (_d.node.filtered = !states.allSelected));
                    } else if (isSankeyChart && currentLegendPoint !== d) {
                        d.disabled = !arrayIncludes(sankeyDependentPoints, d);
                    } else {
                        d.disabled = !states.allSelected;
                    }

                    updateData(d, j, chartObj);
                    legend_helpers.toggleChartEvents(chartObj, true);
                    legend_helpers.fadeChart(true, chartObj, disabledOpacity, true, order);
                    chartObj.eventHandler.tooltipElement.hide();
                }

                if (isSankeyChart && currentLegendPoint !== d && legendFilterAllowed) {
                    if (enabling) {
                        if (d.disabled) {
                            d.disabled = !arrayIncludes(sankeyDependentPoints, d);
                        }
                    } else {
                        if (!d.disabled) {
                            d.disabled = arrayIncludes(sankeyEmptyNodes, d);
                        }
                    }
                }

                return !isHiddenItem(d) && isItemDisabled(d);
            }),
            disableCheck = i => {
                var d = legData[i];
                return !isItemDisabled(d);
            },
            activeCheck = i => {
                var d = legData[i];
                return !isHiddenItem(d);
            },
            activeItems = indices.filter(activeCheck);

        return disabledEle.length == legData_length - activeItems.length && activeItems.every(disableCheck);
    }

    /**
     *
     * @param point
     * @return {boolean}
     */
    function allowLegendFilter(point) {
        // #ZC2653 allow legend click after first click after select all deselection
        if (!states.allSelected && !states.isFirstClickAfterDeselection) {
            return true;
        } else if (isSankeyChart && !point.disabled) {
            const sankeyLayout = chartObj.dataObject.sankeyLayout;
            let node = chartObj.renderer[0].getSankeyNodeByPoint(point),
                linkLength = chartObj.dataObject.sankeyLayout.linkLength;

            const sankeyAlign = chartObj.renderer[0].sankeyOptions.align;
            if (sankeyAlign == 'level') {
                let visibleNodesInLayer = false;
                const nodesInLayer = sankeyLayout.layout.matrix.nodes[node.layer];
                for (let i = 0, _node; i < nodesInLayer.length, _node = nodesInLayer[i]; i++) {
                    if(_node.point.disabled !== true && _node.index !== node.index) {
                        visibleNodesInLayer = true;
                        break;
                    }
                }
                return visibleNodesInLayer;
            }   

            if (chartObj.systemConf.legend.ranges) {
                const nodesInLayer = chartObj.dataObject.sankeyLayout.nodeMatrix[node.layer];
                let legendFilterAllowed = false;
                legendFilterAllowed = (nodesInLayer.find(_node => {
                    if (_node.index !== node.index) {
                        legendFilterAllowed = (_node.sourceLinks.length > 0) ? (_node.sourceLinks.find(_link => _link.point._visible === true) !== undefined) : true;
                        legendFilterAllowed = (_node.targetLinks.length > 0) ? (_node.targetLinks.find(_link => _link.point._visible === true) !== undefined) : legendFilterAllowed;
                        return legendFilterAllowed;
                    }
                }) !== undefined);
                return legendFilterAllowed;
            }
            const totalConnectedLinks = node.sourceLinks.length + node.targetLinks.length;
            
            return linkLength > totalConnectedLinks;
        }
        return true;
    }

    function updateSelectAllBox() {
        if (!states.allSelected) {
            states.allSelected = !states.allSelected;
        }

        selectAllFilter = selectAllFilter || getLegendContainer(chartObj, `clr-${order}`).selectAll('div.selectAll, div.zc-header-legendfilterbox');

        if (!selectAllFilter.empty()) {
            initLegendData(); // ZC1435
            selectAllFilter.html(
                isAllEnabled(legData, states.allSelected) ? colorBoxOptions.text || '&#10003;' : (disabledOptions.colorBox || {}).text || ''
            );
        }
    }

    /**
     *
     * Manually disabling the legend filter for this milestone
     * @return {boolean|*}
     */
    function isLegendFilterEnabled() {
        return legendOptions.filter.enabled;
    }

    function getDistinctAxes(indices) {
        if (chartObj.ismultiy) {
            var set = d3_set();
            return indices.filter(i => {
                var dt = legData[i],
                    key = DataProcessor.helpers.getParsedYAxisOrder(dt.yaxiscolumnorder);

                if (set.has(key)) {
                    return false;
                } else {
                    set.add(key);
                    return true;
                }
            });
        } else {
            return splat(indices[0]);
        }
    }

    function getLegendHighlightComp() {
        if (isFacet(chartObj)) {
            return chartObj.legends.get(`clr-${order}`).highlightManager || chartObj.legendHighlight;
        }

        return chartObj.legendHighlight;
    }
    return legendComponent;
};

export var _legendEvents = legend.events;
