// @zohocharts/zc/src/legend/legend-coloraxis.js
//$Id$
//legend-coloraxis.js
import { MINMAXLABEL_LEGEND_CLASSNAME } from './legend-constants';
import { getLegendAlignment } from './legend-utils';
import { legend_helpers } from '../core/main-zc-init';
import { validateNullVal, bbox, simpleClone } from '../utils/zcutils';
import { TOP_ALIGN, BOTTOM_ALIGN, LEFT_ALIGN, RIGHT_ALIGN } from '../core/align-constants';
import { d3_select, d3_max } from '../core/d3-constants';
import { getDashArray, getLineCap } from '../utils/dashstyles';
import { doOverlap } from '../utils/geometry-utils';
import { addSVGtitleEle, trimText } from '../utils/zcsvgutils';
import { mathMax, mathMin, NONE } from '../core/zc-constants';
import { getSvgTransform } from '../utils/svg-transforms';

function handleOverlapNOverlap(axisGroup, axisGroupPosition, position, chartObj, spec, colorAxisOptions, container, _m) {
    var minmaxlabels = _m || axisGroup.selectAll(`g.${MINMAXLABEL_LEGEND_CLASSNAME}`).selectAll('text'),
        ticklabels = axisGroup.selectAll('g.tick').selectAll('text'),
        minmaxlabelsBoundArray = handleOverflowNGetBoundForMinmaxlabels(minmaxlabels, axisGroupPosition, position, chartObj, spec);

    handleOverflowNOverlapForTicklabels(ticklabels, axisGroupPosition, position, minmaxlabelsBoundArray, chartObj, spec, container);

    if (colorAxisOptions.ticklabel.tooltip === true) {
        //#ZC1327 // Have to enhance to show formatted value and user specified labels
        ticklabels.each(function(d) {
            addSVGtitleEle(d3_select(this), this.__formattedValue);
        });
    }
}

function handleOverflowNGetBoundForMinmaxlabels(minmaxlabels, axisGroupPosition, position, chartObj, spec) {
    var { config, bound } = spec,
        legendOptions = config,
        legendAlign = getLegendAlignment(legendOptions),
        minmaxlabelsBoundArray = [],
        bound = getAvailableLegendBound(config, bound, chartObj.chartarea, chartObj.canvasarea),
        edge = {
            left: bound.x1,
            top: bound.y1,
            right: bound.x2,
            bottom: bound.y2
        },
        margin = {
            left: 5,
            right: 5,
            top: 5,
            bottom: 5
        },
        pattern = '..';

    minmaxlabels.each(function(d, i) {
        var pos = getMinmaxLabelBound(this, axisGroupPosition, position);

        var textBound = {
            left: pos.x1,
            top: pos.y1,
            width: pos.width,
            height: pos.height
        };
        trimText(this, this.textContent, textBound, edge, BOTTOM_ALIGN, pattern, margin);
        if (this.textContent === pattern) {
            d3_select(this).text('');
        }
        minmaxlabelsBoundArray[i] = getMinmaxLabelBound(this, axisGroupPosition, position);
    });

    if (legendAlign.layout === 'horizontal') {
        if (minmaxlabelsBoundArray.length && doOverlap(minmaxlabelsBoundArray[0], minmaxlabelsBoundArray[1])) {
            var bandWidth = minmaxlabelsBoundArray[1].x2 - minmaxlabelsBoundArray[0].x1;
            minmaxlabels.each(function(d, i) {
                var pos = getMinmaxLabelBound(this, axisGroupPosition, position);
                var textBound = {
                    left: pos.x1,
                    top: pos.y1,
                    width: pos.width,
                    height: pos.height
                };
                var e = simpleClone(edge);
                if (i === 0) {
                    e.right = pos.x1 + bandWidth / 2; //need to change
                } else {
                    e.left = pos.x2 - bandWidth / 2; //need to change
                }

                trimText(this, this.textContent, textBound, e, BOTTOM_ALIGN, pattern, margin);
                if (this.textContent === pattern) {
                    d3_select(this).text('');
                }
                minmaxlabelsBoundArray[i] = getMinmaxLabelBound(this, axisGroupPosition, position);
            });
        }
    }

    return minmaxlabelsBoundArray;
}

function getMinmaxLabelBound(element, axisGroupPosition, position, parentPos) {
    parentPos = parentPos || [0, 0];
    var bb = bbox(element);
    var pos = {
        x1: bb.x + axisGroupPosition[0] + position[0] + parentPos[0],
        y1: bb.y + axisGroupPosition[1] + position[1] + parentPos[1],
        width: bb.width,
        height: bb.height
    };
    pos.x2 = pos.x1 + bb.width;
    pos.y2 = pos.y1 + bb.height;
    return pos;
}

function handleOverflowNOverlapForTicklabels(ticklabels, axisGroupPosition, position, minmaxlabelsBoundArray, chartObj, spec, container) {
    var legBBox = bbox(container.node()),
        ticklabelsBoundArray = [],
        maxTicklabelWidth = [],
        count = 0;

    ticklabels.each(function() {
        maxTicklabelWidth.push(bbox(this).width);
    });
    maxTicklabelWidth = d3_max(maxTicklabelWidth);

    ticklabels.each(function() {
        var parentPos = getSvgTransform(this.parentNode).translate;
        ticklabelsBoundArray.push(getMinmaxLabelBound(this, axisGroupPosition, position, parentPos));

        if (handleOverflow(this, chartObj, parentPos, ticklabelsBoundArray[count], maxTicklabelWidth, legBBox, spec, container)) {
            ticklabelsBoundArray[count] = getMinmaxLabelBound(this, axisGroupPosition, position, parentPos);
        }

        if (handleOverlap(this, minmaxlabelsBoundArray, ticklabelsBoundArray, count)) {
            ticklabelsBoundArray[count] = {
                x1: 0,
                y1: 0,
                x2: 0,
                y2: 0
            };
        }

        count++;
    });
}

function handleOverlap(element, minmaxlabelsBoundArray, ticklabelsBoundArray, i) {
    var overlap;

    for (var index = i - 1; index >= 0 && !overlap; index--) {
        overlap = doOverlap(ticklabelsBoundArray[index], ticklabelsBoundArray[i]);
    }

    if (
        (minmaxlabelsBoundArray.length
            ? doOverlap(ticklabelsBoundArray[i], minmaxlabelsBoundArray[0]) || doOverlap(ticklabelsBoundArray[i], minmaxlabelsBoundArray[1])
            : false) ||
        overlap
    ) {
        d3_select(element.parentNode).style('display', NONE);
        return true;
    }
}

function handleOverflow(element, chartObj, parentPos, bb, maxTicklabelWidth, legBBox, spec, container) {
    var { config, bound } = spec,
        orient = legend_helpers.getOrient(config),
        percent = orient === RIGHT_ALIGN || orient === LEFT_ALIGN ? 1 : 0.5,
        bound = getAvailableLegendBound(config, bound, chartObj.chartarea, chartObj.canvasarea),
        legendtitleWidth = config.title.show ? bbox(container.selectAll('.zc-legend-title > text').node()).width : 0,
        w = (bound.x2 - bound.x1 - (legBBox.width - mathMax(maxTicklabelWidth, legendtitleWidth))) * percent,
        h = bound.y2 - bound.y1,
        edge = {
            left: parentPos[0],
            top: parentPos[1],
            right: parentPos[0] + w,
            bottom: parentPos[1] + h
        },
        margin = {
            left: 5,
            right: 5,
            top: 5,
            bottom: 5
        },
        pattern = '..';

    if (bb.x2 - bb.x1 > w) {
        var textBound = {
            left: parentPos[0],
            top: parentPos[1],
            width: bb.width,
            height: bb.height
        };
        trimText(element, element.textContent, textBound, edge, BOTTOM_ALIGN, pattern, margin);
        if (element.textContent === pattern) {
            d3_select(element).text('');
            d3_select(element.parentNode).style('opacity', 0);
        }

        return true;
    }
}

function getAvailableLegendBound(config, bound, chartarea, canvasarea) {
    var legendOptions = config,
        orient = legend_helpers.getOrient(legendOptions),
        legendAlign = getLegendAlignment(legendOptions),
        isHorizontalLayout = legendAlign.layout === 'horizontal',
        hAlign = legendAlign.hAlign,
        vAlign = legendAlign.vAlign,
        floating = legendAlign.floating,
        { width: chartAreaWidth, height: chartAreaHeight, left: chartAreaLeft, top: chartAreaTop } = chartarea,
        w = chartAreaWidth - legendOptions.marginLeft - legendOptions.marginRight,
        h = chartAreaHeight - legendOptions.marginTop - legendOptions.marginBottom,
        l = chartAreaLeft + legendOptions.marginLeft,
        t = chartAreaTop + legendOptions.marginTop,
        { maxWidth, maxHeight } = bound,
        bound = {};

    if (floating) {
        bound.x1 = mathMin(legendOptions.x, chartAreaWidth - maxWidth); // If x or y out of chartarea, move position within chartarea
        bound.y1 = mathMin(legendOptions.y, chartAreaHeight - maxHeight);
        bound.x2 = mathMin(bound.x1 + maxWidth, chartAreaWidth);
        bound.y2 = mathMin(bound.y1 + maxHeight, chartAreaHeight);

        return bound;
    }
    if (orient === BOTTOM_ALIGN) {
        bound.y1 = t + h - maxHeight;
        if (hAlign === LEFT_ALIGN) {
            bound.x1 = l;
        } else if (hAlign === RIGHT_ALIGN) {
            bound.x1 = l + w - maxWidth;
        } else {
            bound.x1 = l + (w - maxWidth) / 2;
        }
    } else if (orient === TOP_ALIGN) {
        bound.y1 = t;
        if (hAlign === LEFT_ALIGN) {
            bound.x1 = l;
        } else if (hAlign === RIGHT_ALIGN) {
            bound.x1 = l + w - maxWidth;
        } else {
            bound.x1 = l + (w - maxWidth) / 2;
        }
    } else if (orient === RIGHT_ALIGN) {
        bound.y1 = t;
        bound.x1 = l + w - maxWidth;
    } else {
        bound.y1 = t;
        bound.x1 = l;
    }

    bound.x2 = bound.x1 + maxWidth;
    bound.y2 = bound.y1 + maxHeight;
    return bound;
}

function applyAxislineProperties(element, axislineOptions, orient, isHorizontal, dimension) {
    if (!axislineOptions.show) {
        element.remove();
    } else {
        var strokeWidth = validateNullVal(axislineOptions.strokeWidth, 1),
            tx = 0,
            ty = 0;

        if (orient === LEFT_ALIGN) {
            tx = strokeWidth / 2;
        } else if (orient === RIGHT_ALIGN) {
            tx = -strokeWidth / 2;
        } else if (orient === BOTTOM_ALIGN) {
            ty = -strokeWidth / 2;
        } else {
            ty = strokeWidth / 2;
        }
        element
            .attr('transform', `translate(${tx},${ty})`)
            .attr('d', isHorizontal ? `M0,0 L${dimension[0]},0` : `M0,0 L0,${dimension[1]}`)
            .styles({
                'stroke-dasharray': getDashArray(axislineOptions.style, strokeWidth),
                'stroke-linecap': getLineCap(axislineOptions.lineCap, axislineOptions.style)
            });
    }
}

export { applyAxislineProperties, handleOverlapNOverlap };
