// @zohocharts/zc/src/legend/legend-colorbox.js
//$Id$
//legend-colorbox.js

import { legend, legend_helpers } from '../core/main-zc-init';
// import { legend_helpers } from './legend-helper';
import { legutil } from './legend-utils';
import { utils, defined, pick, generateIdByPropValue } from '../utils/zcutils';
import { $Browser } from '../core/browser-detect';
import { getColorByIndex, getGrayScaleColor, getColorByScale } from '../utils/colorutils';
import { effects, getChartEffect } from '../effects/zceffects';
import { addPatternThemeByEle, patternThemes, getPattern } from '../effects/patternThemes';
import { getSVGSymbolString, getHTML4SVGSymbol } from '../utils/svg-symbols';
import { colorscale_helpers } from '../axis/colorscale-helpers';
import { mathMax, PX } from '../core/zc-constants';
import { getDisabledOptions } from './events/legend-event-helpers';


var getClr = legend_helpers.getClr,
    isDisabled = legend_helpers.isDisabled,
    hexToRGB = legend_helpers.hexToRGB;

legend.setColorBoxBackground = function(elements, chartObj, states, order, meta) {
    var legendOptions =
            order > 1 && chartObj.systemConf.legend.color ? chartObj.systemConf.legend.color[order] : chartObj.systemConf.legend,
        colorBoxOptions = legendOptions.colorBox,
        disabledOptions = getDisabledOptions(legendOptions, 'clr'),
        disabledColorBoxOptions = disabledOptions.colorBox || {},
        //current Pallete priority
        //imagePallete > patternPallete > colors
        imagePallete = legendOptions.imagePallete ? legutil.imagePallete.get(legendOptions.imagePallete) : null,
        useImagePallete = chartObj.imagePallete && defined(imagePallete) && imagePallete.length > 0,
        patternPallete = chartObj.patternPallete,
        patternPalleteLength = patternPallete && patternPallete.length,
        useSVGImage = !chartObj.patternPallete && (defined(legendOptions.shapes) || colorBoxOptions.shape !== 'square'),
        effect = getChartEffect(legendOptions, chartObj),
        grayScale = disabledColorBoxOptions.grayScale,
        fontSize = +utils.getVal(legendOptions.fontSize || chartObj.systemConf.canvas.fontSize),
        colorBoxSize = legend_helpers.getFilterBoxSize(colorBoxOptions, fontSize),
        colorBoxWidth = colorBoxSize[0],
        colorBoxHeight = colorBoxSize[1],
        background,
        backgroundColor,
        borderRadius = colorBoxOptions.borderRadius,
        backgroundSizeArray = [],
        scaleComp = meta && meta.parserScaleComp,
        _clr = (d, i, _default, order) => {
            //getColorByScale(chartObj, d, scaleComp, column.dataindex)
            //need to pass column.dataindex
            let legendColor = scaleComp ? getColorByScale(chartObj, d, scaleComp, 0) : getColorByIndex(chartObj, i, null, order),
                _ = getClr(_disabled(d), pick(_default, legendColor), disabledOptions);
            if ($Browser.IE && grayScale && _disabled(d)) {
                _ = getGrayScaleColor(_);
            }
            return _;
        },
        _disabled = d => isDisabled(d, states.allSelected, states.isFirstClickAfterDeselection),
        _borderColor = (d, i) => _clr(d, i, colorBoxOptions.strokeColor, order),
        _background = (d, i) => {
            let disabled = _disabled(d);
            let backgroundSize = null;

            backgroundColor = _clr(d, i, colorBoxOptions.fillColor, order);

            if (useImagePallete) {
                if (disabled && disabledOptions.color) {
                    background = disabledOptions.color;
                } else {
                    background = `url("${imagePallete[i % imagePallete.length]}")`;
                    var patternSize = chartObj.patternImageSizes && chartObj.patternImageSizes;
                    backgroundSize = patternSize[i]
                        ? `${patternSize[i % patternSize.length][0] + PX} ${patternSize[i % patternSize.length][1]}${PX}`
                        : null;
                }
            } else if (useSVGImage) {
                background = getsvgImage(colorBoxWidth, colorBoxHeight, _borderColor(d, i), backgroundColor, chartObj, i, disabled);
            } else if (patternPalleteLength) {
                var svgPattern = patternPallete[i % patternPalleteLength];
                var prop = patternThemes[svgPattern] ? getPattern(svgPattern) : addPatternThemeByEle(svgPattern);
                prop.bgClr_base = backgroundColor;
                prop.stroke_base = backgroundColor;
                prop.width = colorBoxWidth;
                prop.customStaticClr = disabled ? disabledOptions.patternColor : chartObj.userdata.legend.customPatternColor;
                prop.x = prop.y = 0;
                var patternID = `zp${generateIdByPropValue(prop, 'legelegendndFill')}`;
                var p = effects.svg_patterns.svgPatternFill(
                    prop,
                    patternID,
                    null,
                    true,
                    true,
                    chartObj,
                    chartObj.getPatternSvg(),
                    'objectBoundingBox'
                );
                var patternWidth = p.width,
                    patternHeight = p.height;
                background = `url("${p.url}")`;
                backgroundSize = `${(patternWidth * 2 < colorBoxWidth ? patternWidth : colorBoxWidth / 2) + PX} ${patternHeight * 2 < colorBoxHeight ? patternHeight : colorBoxHeight / 2
                    }${PX}`;
            } else {
                background = backgroundColor;
            }

            if (isImageUrl(background)) {
                backgroundSize = backgroundSize || `${colorBoxWidth + PX} ${colorBoxHeight}${PX}`;
                if (disabled && grayScale && $Browser.IE) {
                    var dim = backgroundSize.split('px').map(d => +d);
                    background = setGrayScaleForIE(background, dim[0], dim[1]);
                }
            }

            backgroundSizeArray[i] = backgroundSize;
            // pushLegendInfo({ object: legendInfoObj_colorBox.seriesSpecificStyles, key: i, value: { background: background } });
            return background;
        },
        _styles = {
            opacity: (d, i) => {
                var _opacity = _disabled(d) ? disabledColorBoxOptions.fillOpacity : null;
                // pushLegendInfo({ object: legendInfoObj_colorBox.seriesSpecificStyles, key: i, value: { opacity: _opacity } });
                return _opacity;
            },
            filter: (d, i) => {
                let _effects = [];
                _effects.push(_disabled(d) && grayScale && !$Browser.IE ? 'grayscale(1)' : null);
                _effects.push(effect);
                _effects = _effects.filter(_ => !!_);
                var _filter = _effects.length ? _effects.join(' ') : null;
                // pushLegendInfo({ object: legendInfoObj_colorBox.seriesSpecificStyles, key: i, value: { filter: _filter } });
                return _filter;
            },
            'border-radius': !useSVGImage && defined(borderRadius) ? borderRadius + PX : null,
            color: (d, i) => {
                var _color = _disabled(d) ? disabledColorBoxOptions.fontColor : colorBoxOptions.fontColor || 'white';
                // pushLegendInfo({ object: legendInfoObj_colorBox.seriesSpecificStyles, key: i, value: { color: _color } });
                return _color;
            },
            border: (d, i) => {
                var _border = useImagePallete || useSVGImage ? null : `${colorBoxOptions.strokeWidth + PX}solid ${_borderColor(d, i)}`;
                // pushLegendInfo({ object: legendInfoObj_colorBox.seriesSpecificStyles, key: i, value: { border: _border } });
                return _border;
            },
            background: _background,
            'background-size': (d, i) => {
                // pushLegendInfo({
                //     object: legendInfoObj_colorBox.seriesSpecificStyles,
                //     key: i,
                //     value: { backgroundSize: backgroundSizeArray[i] }
                // });
                return backgroundSizeArray[i];
            }
        },
        _html = (d, i) => {
            var content = _disabled(d) ? disabledColorBoxOptions.text || '' : colorBoxOptions.text;
            content = typeof content === 'string' ? content.replace(/<|>/g, _ => _ === '<' ? '&lt;' : _ === '>' ? '&gt;' : _) : content;
            // pushLegendInfo({ object: legendInfoObj_colorBox.seriesSpecificStyles, key: i, value: { html: content } });
            return content;
        };

    // var legendInfoObj = chartObj.legend.unitTestObject['zc-legend'].clr[order],
    //     legendInfoObj_colorBox = legendInfoObj.legend_grp.legend_nodes.colorBox;

    // pushLegendInfo({
    //     object: legendInfoObj_colorBox,
    //     value: {
    //         borderRadius: !useSVGImage && defined(borderRadius) ? borderRadius : null
    //     }
    // });

    if (elements) {
        elements.styles(_styles).html(_html);
    } else {
        return {
            styles: _styles,
            html: _html
        };
    }
};

function getsvgImage(width, height, strokeColor, fillColor, chartObj, index, disabled) {
    //SVG data uri as div background
    strokeColor = hexToRGB(strokeColor);
    fillColor = hexToRGB(fillColor);
    var legendOptions = chartObj.systemConf.legend,
        colorBoxOptions = legendOptions.colorBox,
        symbolsArray = legendOptions.shapes,
        symbol = (symbolsArray ? symbolsArray[index % symbolsArray.length] : null) || colorBoxOptions.shape;

    var elements =
        !symbol &&
            legend.getHTML4ChartTypeSymbol &&
            !(colorscale_helpers.isQuantileColorScale(chartObj) && chartObj.dataObject.isCombinationalChart)
            ? legend.getHTML4ChartTypeSymbol(width, height, fillColor, chartObj, index, disabled)
            // : getHTML4SVGSymbol(width, height, strokeColor, fillColor, symbol, colorBoxOptions);
            : getHTML4SVGSymbol({ ...colorBoxOptions, width: width, height: height, strokeColor: strokeColor, fillColor: fillColor, shape: symbol });
    if (isImageUrl(elements)) {
        return `url("${elements}")`;
    }
    return getSVGSymbolString(elements, width, height);
}

function isImageUrl(str) {
    return str && !(str[0] === '<' || str.indexOf('url("data') === 0);
}

function setGrayScaleForIE(src, w, h) {
    // www.majas-lapu-izstrade.lv/cross-browser-grayscale-image-example-using-css3-js/
    var canvas = document.createElement('canvas');
    var ctx = canvas.getContext('2d');
    canvas.width = w;
    canvas.height = h;

    var imgObj = new Image();
    imgObj.src = src.substring(5, src.length - 2); // (e.g): src = url("<<URL>>")
    imgObj.width = w;
    imgObj.height = h;

    imgObj.onload = function () {
        ctx.drawImage(imgObj, 0, 0, w, h);
        var imgPixels = ctx.getImageData(0, 0, canvas.width, canvas.height);
        for (var y = 0; y < imgPixels.height; y++) {
            for (var x = 0; x < imgPixels.width; x++) {
                var i = y * 4 * imgPixels.width + x * 4;
                var avg = (imgPixels.data[i] + imgPixels.data[i + 1] + imgPixels.data[i + 2]) / 3;
                imgPixels.data[i] = avg;
                imgPixels.data[i + 1] = avg;
                imgPixels.data[i + 2] = avg;
            }
        }
        ctx.putImageData(imgPixels, 0, 0, 0, 0, imgPixels.width, imgPixels.height);
        return `url(${canvas.toDataURL()})`;
    };
}

// /**
//  *
//  * @param width
//  * @param height
//  * @param strokeColor
//  * @param fillColor
//  * @param symbol
//  * @param colorBoxOptions
//  * @param dontScale
//  * @param {number | string} [rotate = 0] - rotation in deg | turn | rad. ***default = 0***
//  * @return {string}
//  */
// export function getHTML4SVGSymbol(width, height, strokeColor, fillColor, symbol, colorBoxOptions, dontScale, rotate = 0) {
//     var { fillOpacity = 1, strokeOpacity = 1, strokeWidth, padding = 0, borderRadius } = colorBoxOptions,
//         size_w = dontScale ? mathMax(width - padding * 2, 1) : 10,
//         size_h = dontScale ? mathMax(height - padding * 2, 1) : 10,
//         _wF = mathMax(width - padding * 2, 1) / (size_w + 2 * validateNullVal(strokeWidth, 1)),
//         _hF = mathMax(height - padding * 2, 1) / (size_h + 2 * validateNullVal(strokeWidth, 1)),
//         symbolpath = getSVGsymbol(symbol || 'square')(size_w, size_h, borderRadius),
//         hexToRGB = legend_helpers.hexToRGB;

//     return `<path d='${symbolpath}' transform='translate(${width / 2},${height / 2}) scale(${_wF},${_hF}) rotate(${rotate})' style='stroke: ${hexToRGB(
//         strokeColor
//     )}; fill: ${hexToRGB(fillColor)}; fill-opacity: ${fillOpacity}; stroke-width: ${validateNullVal(
//         strokeWidth,
//         1
//     )}px; stroke-opacity: ${strokeOpacity};'></path>`;
// }

// /**
//  * Function that takes configuration of the symbol as parameter.
//  * @param {Object} symbolConfig
//  * @param {number} symbolConfig.size - size of the symbol/shape.
//  * @param {string} symbolConfig.shape - shape of the object
//  * @param {number} symbolConfig.width 
//  * @param {number} symbolConfig.height
//  * @param {string} symbolConfig.fillColor
//  * @param {number} symbolConfig.fillOpacity
//  * @param {string} symbolConfig.strokeColor
//  * @param {number} symbolConfig.strokeWidth
//  * @param {number} symbolConfig.strokeOpacity
//  * @param {number} symbolConfig.padding
//  * @param {number} symbolConfig.borderRadius
//  * @param {number} symbolConfig.rotate - rotation in deg | turn | rad. ***default = 0***
//  * @param {boolean} dontScale 
//  * @returns {string} svg path in string format
//  */

// export function getHTML4SVGSymbol(symbolConfig, dontScale) {
//     const { size = 1, width = symbolConfig.width || size, height = symbolConfig.height || size, shape, fillColor = 'black',
//         fillOpacity = 1, strokeColor = 'black', strokeWidth = 1, strokeOpacity = 1, padding = 0, borderRadius = 0, rotate = 0,
//         translateX = symbolConfig.translateX || (width / 2), translateY = symbolConfig.translateY || (height / 2) } = symbolConfig,
//         size_w = dontScale ? mathMax(width - padding * 2, 1) : 10,
//         size_h = dontScale ? mathMax(height - padding * 2, 1) : 10,
//         _wF = mathMax(width - padding * 2, 1) / (size_w + 2 * validateNullVal(strokeWidth, 1)),
//         _hF = mathMax(height - padding * 2, 1) / (size_h + 2 * validateNullVal(strokeWidth, 1)),
//         symbolpath = getSVGsymbol(shape || 'square')(size_w, size_h, borderRadius),
//         hexToRGB = legend_helpers.hexToRGB;

//     return `<path d='${symbolpath}' transform='translate(${translateX},${translateY}) scale(${_wF},${_hF}) rotate(${rotate
//         })' style='stroke: ${hexToRGB(strokeColor)}; fill: ${hexToRGB(fillColor)}; fill-opacity: ${fillOpacity
//         }; stroke-width: ${validateNullVal(strokeWidth, 1)}px; stroke-opacity: ${strokeOpacity};'></path>`;
// }

// /**
//  *
//  * @param pathString
//  * @param width
//  * @param height
//  * @param {boolean} notAsURL
//  * @return {string}
//  */

// // export function getSVGSymbolString(pathString, width, height) {
// //     var xml = `<svg xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' width='${width}' height='${height}'>`;
// //     xml += pathString;
// //     xml += ' </svg>';
// //     return `${'url("' + 'data:image/svg+xml;'}${$Browser.IE ? 'base64' : 'utf8'},${$Browser.IE ? btoa(xml) : xml}")`;
// // }

// export function getSVGSymbolString(pathString, width, height, notAsURL) {
//     var xml = `<svg xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' width='${width}' height='${height}'>`;
//     xml += pathString;
//     xml += ' </svg>';
//     let url = `data:image/svg+xml;'}${$Browser.IE ? 'base64' : 'utf8'},${$Browser.IE ? btoa(xml) : xml}`;
//     return notAsURL ? url : `url("${url}")`;
// }

/**
 *
 * @param vAlign
 * @param labelHeight
 * @param colorBoxHeight
 * @param fontSize
 * @param colorBoxSize
 * @param textOverflow
 * @return {number}
 */
export function getColorBoxVerticalPosition(vAlign, labelHeight, colorBoxHeight, fontSize, colorBoxSize, textOverflow) {
    let textHangingGap = ~~(fontSize * 1.195 - fontSize);
    var margin_top = 0;

    switch (vAlign) {
        case 'top': {
            margin_top = textHangingGap;
            break;
        }
        case 'bottom': {
            margin_top = labelHeight - colorBoxHeight - textHangingGap;
            break;
        }
        default: {
            if (vAlign === 'center' && textOverflow !== 'wrap') {
                margin_top = (mathMax(labelHeight, colorBoxHeight) - colorBoxHeight) / 2;
            } else if (defined(colorBoxSize) || textOverflow === 'wrap') {
                margin_top = (mathMax(fontSize, colorBoxHeight) - colorBoxHeight) / 2;
            } else if (labelHeight > colorBoxHeight) {
                margin_top = (labelHeight - colorBoxHeight) / 2;
            }
        }
    }
    return margin_top;
}
