// @zohocharts/zc/src/legend/legend-continuous.js
//$Id$
//legend-continuous.js
import { TRANSPARENT } from '../core/zc-constants';
import { mergeJSON } from '../utils/zcutils';

function getColorAxisOptions(legendOptions) {
    var defaultColorAxisOptions = {
        label: {},
        grid: {
            color: TRANSPARENT
        },
        ticklabel: {
            fontColor: legendOptions.fontColor,
            fontFamily: legendOptions.fontFamily,
            fontSize: legendOptions.fontSize,
            fontStyle: legendOptions.fontStyle,
            fontWeight: legendOptions.fontWeight,
            // marginBottom: 5,
            // marginLeft: 5,
            // marginRight: 5,
            // marginTop: 5,
            // maxHeight : "",
            // maxWidth : "",
            // rotation: 0,
            // textOverflow: "ellipsis",
            tooltip: legendOptions.tooltip
        },
        minmaxlabels: {
            //excludeIntervals, show, padding, align, text, fontProperties
            padding: 2,
            align: 'center',
            fontSize: legendOptions.fontSize
        },
        tickmark: {
            color: '#B0B0B0',
            size: 3
        },
        axisline: {
            show: false
        }
    };
    var colorAxisOptions = legendOptions.colorAxis || {};
    colorAxisOptions = mergeJSON(defaultColorAxisOptions, colorAxisOptions, true);
    return colorAxisOptions;
}

export { getColorAxisOptions };
