// @zohocharts/zc/src/legend/legend-discrete.js
//$Id$
//legend-discrete.js
import { ExpandableAdjustFactor } from './expandable/expandable-adjust';
import { legend, legend_helpers } from '../core/main-zc-init';
import { d3_dispatch, d3_rgb, d3_select } from '../core/d3-constants';
import {
    decodeHTML,
    defined,
    elementIsScrollable,
    getBoundingRect,
    getValByPriority,
    perToPx,
    pick,
    shorthand_px,
    simpleClone,
    utils,
    invokeFunction
} from '../utils/zcutils';
import { getLegendAlignment } from './legend-utils';
import { appendEle, applyFont, createElement } from '../utils/zcsvgutils';
import { isFunction } from '../utils/function-utils';
import { mathFloor, mathMax, mathMin, mathRound, NONE, POINTER, PX, TRANSPARENT } from '../core/zc-constants';
import { $Browser } from '../core/browser-detect';
import { getLegendPosition } from './legend-utils';
import {
    LEGEND_TITLE_MAX_WIDTH_HORIZONTAL,
    LEGEND_TITLE_MAX_WIDTH_VERTICAL,
    ZC_LEGEND_maxHeightHoriRatio,
    ZC_LEGEND_maxHeightVertRatio,
    ZC_LEGEND_maxWidthHoriRatio,
    ZC_LEGEND_maxWidthVertRatio
} from './legend-constants';
import { colorscale_helpers } from '../axis/colorscale-helpers';
import { $ZCG } from '../core/zcglobal';
import { getColorBoxVerticalPosition } from './legend-colorbox';
import { helpers } from './legend-discrete-helpers';
import { expandableLegend } from './legend-expandable';
import { getDisabledOptions } from './events/legend-event-helpers';
import { isFacet } from '../axis/hierarchy-axis-utils';

legend.discrete = function() {};
legend.discrete.get = function(spec) {
    'use strict';
    var chartarea,
        legendWidth = 0,
        color,
        legendSeriesWidth = 0,
        dispatch = d3_dispatch('legendClick', 'legendDblclick', 'legendMouseover', 'legendMouseout'),
        scrollbarwid = 20,
        wrap,
        systemconf,
        chartObj,
        { bound, globalFont, callbackArgs, order } = spec,
        getLegendData = legend_helpers.getLegendData,
        getLegendDataLength = legend_helpers.getLegendDataLength,
        isHiddenItem = legend_helpers.isHiddenItem,
        getClr = legend_helpers.getClr,
        isDisabled = legend_helpers.isDisabled;

    function chart(selection) {
        selection.each(function(data) {
            chartObj.legend.state = chartObj.legend.state || {};
            var legData = getLegendData(chartObj, order),
                legData_length = getLegendDataLength(chartObj, legData),
                colorBoxOptions = data.colorBox,
                disabledOptions = getDisabledOptions(data, 'clr'),
                priorityArray = [data, systemconf.canvas, $ZCG],
                floating = data.x != null && data.x !== '' && data.y != null && data.y !== '',
                fontSize = +utils.getVal(data.fontSize || systemconf.canvas.fontSize),
                fontColor = getValByPriority(priorityArray, 'fontColor'),
                colIdx = chartObj.dataset.getColumnIdx(
                    chartObj.isAxisCategory ||
                        chartObj.isGeoMap ||
                        chartObj.isNonAxisMultiSeriesCategory ||
                        colorscale_helpers.isQuantileColorScale(chartObj)
                        ? 'clr'
                        : 'x'
                ),
                showAs = data.showAs,
                colorBoxPosition = colorBoxOptions.position || 'left',
                expandable = data.expandable || {},
                legendExpandable = expandable.show === true,
                redrawing = !!chartObj.renderer.length,
                colorBoxSize = legend_helpers.getFilterBoxSize(colorBoxOptions, fontSize),
                colorBoxWidth = colorBoxSize[0],
                colorBoxHeight = colorBoxSize[1],
                legendAlign = getLegendAlignment(data),
                hAlign = legendAlign.hAlign,
                vAlign = legendAlign.vAlign,
                states = (chartObj.legend.state[`clr-${order}`] = chartObj.legend.state[`clr-${order}`] || {
                    allSelected: true,
                    isFirstClickAfterDeselection: false
                }),
                lineHeight = defined(colorBoxOptions.size)
                    ? `${mathMax(fontSize, colorBoxHeight)}px`
                    : ($Browser.OSX || $Browser.LINUX) && $Browser.FIREFOX
                    ? '1.2'
                    : 'normal',
                isColumnView = legendAlign.layout === 'horizontal',
                plotarea = simpleClone(chartObj.plotarea);

            chartObj.legend.states = chartObj.legend.state['clr-0'];

            if (!chartObj.legend.unitTestObject) {
                chartObj.legend.unitTestObject = {
                    'zc-legend': {
                        clr: {}
                    }
                };
            }

            // moved here from inner function
            var ImpFontProp = {
                fontColor: function(d) {
                    return getClr(isDisabled(d, states.allSelected, states.isFirstClickAfterDeselection), fontColor, disabledOptions);
                },
                textDecoration: function(d) {
                    return isDisabled(d, states.allSelected, states.isFirstClickAfterDeselection) && disabledOptions.strikeout
                        ? 'line-through'
                        : null;
                }
            };

            //text-overflow properties
            var item_textOverflow_properties;
            if (data.textOverflow === 'wrap') {
                item_textOverflow_properties = {
                    'word-wrap': 'break-word',
                    'white-space': 'normal',
                    'vertical-align': 'top'
                };
            } else {
                item_textOverflow_properties = {
                    'text-overflow': 'ellipsis',
                    'vertical-align': 'middle'
                };
            }

            var container = d3_select(this);
            chart.update = function() {
                container.transition().call(chart);
            };
            var styleObj = {
                display: 'block',
                position: 'absolute',
                left: null, //Reset left and top position. so that floating legend won't get trimmedon redraw (legend-03.json)
                top: null
            };
            container.styles(styleObj);
            var wrap = appendEle(container, 'div', [data], '.', 'class', 'legendgroup');

            //shadow properties
            var shadow = data.shadow;
            if (shadow.show === true) {
                var shadowstyle = `${shadow.x}px ${shadow.y}px ${shadow.blur}px ${shadow.color}`;
                shadowstyle = shadow.inset ? `${shadowstyle} inset` : shadowstyle;
                container.style('box-shadow', shadowstyle);
            } else {
                container.style('box-shadow', null);
            }

            //Background properties
            var rgb = d3_rgb(data.background.color.toString());
            container.style('background-color', `rgba(${rgb.r},${rgb.g},${rgb.b},${data.background.alpha})`);

            //Border Properties
            var border = data.border;
            if (border.show === true) {
                container.styles({
                    border: `${+border.size}px ${border.style} ${border.color}`,
                    'border-radius': `${border.radius}px`
                });
            } else {
                container.style('border', null);
            }

            //Legend title
            var title_data = data.title,
                keepLegendTitleAligned = title_data.keepAligned && isColumnView,
                titleMargin = defined(title_data.margin) ? title_data.margin : 3,
                legendText = title_data.text,
                titleAlignedStyles = {
                    display: keepLegendTitleAligned ? 'inline-block' : null,
                    'vertical-align': 'top'
                };

            var legendInfoObj = (chartObj.legend.unitTestObject['zc-legend'].clr[order] = {
                fontSize: fontSize,
                fontColor: fontColor,
                background_rgb: rgb,
                title_grp: {
                    title_nodes: {
                        selectAll: {},
                        title: {}
                    }
                },
                legend_grp: {
                    legend_nodes: {
                        seriesSpecificStyles: {},
                        label: {},
                        colorBox: {}
                    }
                }
            });

            var header = createElement('ul', '.zcheader', title_data.show === true ? [title_data] : [], wrap, ':first-child').styles(
                titleAlignedStyles
            );

            var headerLI = appendEle(header, 'li', [1], '.', 'class', 'zcheader');
            header.styles({
                'white-space': 'normal',
                overflow: 'hidden',
                width: '',
                'list-style-type': NONE,
                margin: 0,
                padding: 0
            });
            var headerLI_padding = shorthand_px(titleMargin);
            headerLI.styles({
                padding: headerLI_padding,
                'white-space': 'nowrap',
                overflow: 'hidden',
                'text-align': 'left',
                'text-overflow': 'ellipsis',
                'vertical-align': 'middle'
            });
            pushLegendInfo({ object: legendInfoObj.title_grp, value: { ...titleAlignedStyles } });
            pushLegendInfo({
                object: legendInfoObj.title_grp.title_nodes,
                value: { padding: headerLI_padding }
            });

            var selectAllObj = legend.selectAllBox(headerLI, chartObj, states, order);
            var selectAllFilter = selectAllObj.element,
                selectAllBoxOptions = selectAllObj.options,
                selectAllFilterSize = selectAllObj.size,
                disabledOpacity = selectAllBoxOptions.disabledOpacity;

            if (!isFacet(chartObj)) {
                legend_helpers.fadeChart(states.allSelected, chartObj, disabledOpacity);
            }

            var legtitle = appendEle(headerLI, 'span', [1], '.', 'class', 'legendtitle');
            var padding_attr = `padding-${colorBoxPosition}`,
                padding_val = selectAllBoxOptions.show ? `${data.textPadding}px` : null,
                margin_attr = `margin-${colorBoxPosition}`,
                margin_val = selectAllBoxOptions.show ? `${selectAllFilterSize}px` : null,
                titleContent,
                titleStyles = {};

            legtitle.style(padding_attr, padding_val).style(margin_attr, margin_val);

            if (isFunction(legendText)) {
                titleContent = invokeFunction(legendText, chartObj);
                legtitle.html(titleContent);
            } else {
                titleContent = decodeHTML(legendText);
                legtitle.text(titleContent);
            }
            var titleFontStyles = applyFont(legtitle, [title_data, systemconf.canvas, $ZCG], true);

            titleStyles[padding_attr] = padding_val;
            titleStyles[margin_attr] = margin_val;
            pushLegendInfo({
                object: legendInfoObj.title_grp.title_nodes.title,
                value: {
                    ...titleStyles,
                    html_or_text: titleContent,
                    ...titleFontStyles
                }
            });

            selectAllFilter.style('margin', function() {
                var top = `${Math.max(0, (legtitle.node().offsetHeight - selectAllFilter.node().offsetHeight) / 2)}px`;

                // The bellow lines are commented because of #ZC1538 & #ZC1510
                //     bottom = "0px",
                //     left = ((colorBoxWidth - selectAllFilterSize) / 2) + "px",
                //     right = ((colorBoxWidth - selectAllFilterSize) / 2) + "px";
                //     return top + " " + right + " " + bottom + " " + left;
                var margin = `${top} 0px 0px 0px`;
                pushLegendInfo({
                    object: legendInfoObj.title_grp.title_nodes.selectAll,
                    value: { margin: margin }
                });
                return margin;
            });

            headerLI.style('line-height', function() {
                var selectAllFilterHeight = selectAllBoxOptions.show ? selectAllFilter.node().offsetHeight : 0;
                var headerLineHeight = Math.max(legtitle.node().offsetHeight, selectAllFilterHeight);
                pushLegendInfo({
                    object: legendInfoObj.title_grp.title_nodes,
                    value: { lineHeight: headerLineHeight }
                });
                return `${headerLineHeight}px`;
            });

            var legseries = appendEle(wrap, 'ul', [data], '.', 'class', 'legseries');
            legseries
                .styles({
                    'white-space': 'normal',
                    'overflow-y': 'hidden',
                    // "cursor": cursor,
                    width: '',
                    'list-style-type': NONE,
                    margin: 0,
                    padding: 0,
                    '-webkit-overflow-scrolling': 'touch',

                    // #ZC758
                    // https://stackoverflow.com/questions/5256533/
                    'font-size': '0px'
                })
                .styles(titleAlignedStyles);

            // set the ul to legend discrete helper to work
            // legendDiscreteHelper.setItemParent(legseries);

            // moved code to a function to render the legend list
            const renderLegendList = legData => {
                var series = appendEle(legseries, 'li', legData, '.', 'class', 'series');
                appendEle(series, 'div', [1], '.', 'class', 'legendfilterbox');
                appendEle(series, 'span', [1], '.', 'class', 'legendlabel');

                series
                    .styles({
                        color: function(d, i) {
                            var _color = getClr(
                                isDisabled(d, states.allSelected, states.isFirstClickAfterDeselection),
                                fontColor,
                                disabledOptions
                            );
                            pushLegendInfo({
                                object: legendInfoObj.legend_grp.legend_nodes.seriesSpecificStyles,
                                key: i,
                                value: { color: _color }
                            });
                            return _color;
                        },
                        position: 'relative',
                        'white-space': 'nowrap',
                        padding: `${+data.itemMargin}px`,
                        overflow: 'hidden',
                        'text-align': 'left',
                        'line-height': lineHeight,
                        'font-size': `${fontSize}px`,
                        // width: data.itemWidth == null || data.itemWidth == '' ? null : `${data.itemWidth}px`,
                        width: itemWithIsNumber ? `${data.itemWidth}px` : null,
                        height: data.itemHeight == null || data.itemHeight == '' ? null : `${data.itemHeight}px`,
                        'font-family': getValByPriority(priorityArray, 'fontFamily'),
                        'font-weight': getValByPriority(priorityArray, 'fontWeight'),
                        display: function(d, i) {
                            var _display = isHiddenItem(d) ? NONE : isColumnView ? 'inline-block' : 'block';
                            pushLegendInfo({
                                object: legendInfoObj.legend_grp.legend_nodes.seriesSpecificStyles,
                                key: i,
                                value: { display: _display }
                            });
                            return _display;
                            // return isHiddenItem(d) ? NONE : (isColumnView ? "inline-block" : "inline-block");
                        }
                    })
                    .styles(item_textOverflow_properties)
                    .style('vertical-align', 'top');

                legend.events(chartObj, series, states, disabledOpacity, order);

                var legendlabel = series.select('span.legendlabel'),
                    legendInfoObj_label = legendInfoObj.legend_grp.legend_nodes.label;
                // #ZC1594
                //TODO itemWidth: auto should be supported for all.
                if (defined(showAs)) {
                    var _s = {
                        // #ZC2002
                        display: pick(data.showAsDisplay, 'inline-block'),
                        overflow: 'hidden'
                    };
                    legendlabel.styles(_s);
                    pushLegendInfo({ object: legendInfoObj_label, value: _s });
                    if (defined(chart.legendLabelMaxWidth) && isAutoItemWidth) {
                        legendlabel.style('width', chart.legendLabelMaxWidth + PX);
                        pushLegendInfo({ object: legendInfoObj_label, value: { width: chart.legendLabelMaxWidth } });
                    }
                }

                var padding_attr = `padding-${colorBoxPosition}`,
                    padding_val = data.textPadding,
                    margin_attr = `margin-${colorBoxPosition}`,
                    margin_val = colorBoxWidth,
                    _labelStyle = {};

                legendInfoObj_label.seriesSpecificHTML = {};
                legendlabel
                    .style(padding_attr, `${padding_val}px`)
                    .style(margin_attr, `${margin_val}px`)
                    .html(function(d, i) {
                        var content = legendDiscreteHelper.getLegendItemText(d, i);
                        pushLegendInfo({ object: legendInfoObj_label.seriesSpecificHTML, key: i, value: content });
                        return content;
                    })
                    .attr('title', function() {
                        var title = data.tooltip ? this.textContent : null;
                        pushLegendInfo({ object: legendInfoObj_label, value: { title: title } });
                        return title;
                    });

                var labelFontStyles = applyFont(legendlabel, [ImpFontProp, data, systemconf.canvas, $ZCG], true);

                _labelStyle[padding_attr] = padding_val;
                _labelStyle[margin_attr] = margin_val;
                pushLegendInfo({
                    object: legendInfoObj_label,
                    value: { ..._labelStyle, ...labelFontStyles, color: fontColor, 'text-decoration': null }
                });

                pushLegendInfo({
                    object: legendInfoObj.legend_grp.legend_nodes,
                    value: {
                        lineHeight,
                        width: itemWithIsNumber ? `${data.itemWidth}px` : null,
                        height: data.itemHeight == null || data.itemHeight == '' ? null : `${data.itemHeight}px`,
                        ...item_textOverflow_properties,
                        'font-family': getValByPriority(priorityArray, 'fontFamily'),
                        'font-weight': getValByPriority(priorityArray, 'fontWeight'),
                        'vertical-align': 'top',
                        cursor:
                            (chartObj.systemConf.legend && chartObj.systemConf.legend.events && chartObj.systemConf.legend.events.cursor) ||
                            POINTER
                    }
                });

                legendInfoObj_label.disabledStyles = {};
                // apply disabled showAs styles
                // #ZC1943
                if (defined(showAs)) {
                    series.each(function(d, i) {
                        // #ZC1872 -- apply only if the data is disabled
                        if (isDisabled(d, states.allSelected, states.isFirstClickAfterDeselection)) {
                            var _ = applyFont(
                                d3_select(this).selectAll('.legendlabel, .legendlabel *'),
                                [
                                    { fontColor: ImpFontProp.fontColor(d), textDecoration: ImpFontProp.textDecoration(d) },
                                    data,
                                    systemconf.canvas,
                                    $ZCG
                                ],
                                true
                            );
                            pushLegendInfo({ object: legendInfoObj_label.disabledStyles, key: i, value: _ });
                        }
                    });
                }

                var labelHeight = series.select('span.legendlabel').node().offsetHeight,
                    legendFilter = series.select('div.legendfilterbox'),
                    margin_top = getColorBoxVerticalPosition(
                        colorBoxOptions.vAlign,
                        labelHeight,
                        colorBoxHeight,
                        fontSize,
                        colorBoxOptions.size,
                        data.textOverflow
                    );
                pushLegendInfo({
                    object: legendInfoObj.legend_grp.legend_nodes.colorBox,
                    value: {
                        position: colorBoxPosition,
                        size: colorBoxSize,
                        margin_top,
                        lineHeight: colorBoxHeight - colorBoxOptions.strokeWidth
                    }
                });
                legendFilter.styles({
                    width: colorBoxWidth + PX,
                    height: colorBoxHeight + PX,
                    float: colorBoxPosition,
                    'margin-top': margin_top + PX, // ISSUE:ZC429 -> line-height will not be applied to float element
                    'border-color': TRANSPARENT,
                    'background-repeat': 'repeat',
                    'line-height': colorBoxHeight - colorBoxOptions.strokeWidth + PX,
                    'text-align': 'center',
                    'font-family': colorBoxOptions.fontFamily,
                    'font-size': (colorBoxOptions.fontSize || fontSize) + PX,
                    position: 'absolute',
                    display: 'inline-block',
                    'box-sizing': 'border-box'
                });

                // background for colorBox
                legend.setColorBoxBackground(legendFilter, chartObj, states, order);
                return series;
            };

            // create the helpers
            let legendDiscreteHelper = helpers(
                chartObj,
                legseries,
                legData,
                legData_length,
                colorBoxSize,
                legendAlign,
                lineHeight,
                ImpFontProp
            );

            // title max with depends on the -> calculated legend max width
            // since the legend max width can change based on the layout, so we have put this as a function
            const getTitleMaxWidth = maxWidth => {
                return mathMin(
                    perToPx(
                        title_data.maxWidth ||
                            (keepLegendTitleAligned ? LEGEND_TITLE_MAX_WIDTH_HORIZONTAL : LEGEND_TITLE_MAX_WIDTH_VERTICAL),
                        maxWidth
                    ),
                    maxWidth
                );
            };

            // calcualate the temp height for max number of rows;
            var maxHeight = perToPx(
                    data.maxHeight || (isColumnView ? ZC_LEGEND_maxHeightHoriRatio : ZC_LEGEND_maxHeightVertRatio),
                    chartarea.height
                ),
                maxWidth = perToPx(
                    data.maxWidth || (isColumnView ? ZC_LEGEND_maxWidthHoriRatio : ZC_LEGEND_maxWidthVertRatio),
                    chartarea.width
                );

            if (bound.adjust) {
                maxWidth = mathMin(bound.maxWidth || maxWidth, maxWidth);
                maxHeight -= getBoundingRect(header).height;
                maxHeight = mathMin(bound.maxHeight ? bound.maxHeight - getBoundingRect(header).height : maxHeight, maxHeight);
            }

            var maxWidthTitle = getTitleMaxWidth(maxWidth),
                rowsCount,
                tempDimension = legendDiscreteHelper.getItemDimension(legData[0]),
                tempLabelHeight = tempDimension.label.height,
                tempItemHeight = tempDimension.item.height,
                series = null,
                visibleItemCount = null,
                isAutoItemWidth = data.itemWidth === 'auto',
                itemWithIsNumber = typeof data.itemWidth === 'number',
                // TODO below line issue in bottom horizontal legend
                maxItemHeight = pick(data.itemHeight && +data.itemHeight, mathMax(tempLabelHeight, colorBoxHeight) + 2 * +data.itemMargin),
                headerDimension,
                headerWidth,
                scrollWid,
                constrainedMaxWidth,
                wrapMaxWidth = maxWidth;

            // [#ZC1999] if the max height is less than the item height change the max-height to the item height
            // TODO the same problem will occur in vertical layout and for non expandable legend too
            //  To fix it assign [maxHeight = mathMax(tempItemHeight, maxHeight);] directly with out the condition.
            // if (legendExpandable && isColumnView) {
            //     maxHeight = mathMax(tempItemHeight, maxHeight);
            // }
            if (legendExpandable && isColumnView) {
                maxHeight = mathFloor(maxHeight / tempItemHeight) * tempItemHeight;
                if (maxHeight < tempItemHeight) {
                    maxHeight = tempItemHeight;
                }
            }

            // set if legend is scrollable
            chartObj.legend.hasScroll = legData_length * maxItemHeight > maxHeight;

            // set the max width
            wrap.style('max-width', `${maxWidth}px`);
            header.style('max-width', `${maxWidthTitle}px`);

            headerDimension = getBoundingRect(header);
            headerWidth = headerDimension.width;

            if (isColumnView) {
                // if user has not defined any width then subtract the margin and border
                let extraSpaces = +data.marginLeft + +data.marginRight + (data.border.show ? 2 * data.border.size : 0),
                    allowedLegMaxWidth = chartObj.canvasarea.width - extraSpaces;

                if (maxWidth > allowedLegMaxWidth) {
                    maxWidth = allowedLegMaxWidth;
                }

                // get the new legend maxWidth
                maxWidthTitle = getTitleMaxWidth(maxWidth);

                if (keepLegendTitleAligned) {
                    maxWidth = maxWidth - headerWidth;
                }

                wrap.style('max-width', `${maxWidth}px`);
                header.style('max-width', `${maxWidthTitle}px`);
            }
            pushLegendInfo({ object: legendInfoObj.title_grp, value: { maxWidth: maxWidthTitle } });

            // test cases affected
            if (legendExpandable) {
                rowsCount = legendDiscreteHelper.getRowsCount(tempItemHeight, maxHeight);
                maxHeight = mathMin(rowsCount * tempItemHeight, maxHeight);

                if (bound.adjust) {
                    let count = legendDiscreteHelper.getVisibleItemRenderCount(wrap, rowsCount, maxWidth, maxHeight, tempItemHeight),
                        { maxWidth: maxW, maxHeight: maxH } = bound;

                    if (legData.length !== count) {
                        let expandableAdjust = ExpandableAdjustFactor(expandable, maxW, maxH, globalFont, legendAlign, data.itemMargin);
                        maxHeight -= expandableAdjust[1];
                        maxHeight = mathMin(maxH ? maxH - expandableAdjust[1] : maxHeight, maxHeight);
                    }
                }
            }

            pushLegendInfo({
                object: legendInfoObj.legend_grp,
                value: {
                    maxWidth: maxWidth,
                    maxHeight: maxHeight,
                    ...titleAlignedStyles
                }
            });
            legseries.styles({
                'max-width': maxWidth + PX,
                'max-height': maxHeight + PX
            });
            /**
             * *****************************************************************
             *                      itemWidth: auto                           *
             * *****************************************************************
             */

            // set the max item width possible
            chart.itemMaxWidth = null;
            chart.legendLabelMaxWidth = null;

            if (showAs && isAutoItemWidth) {
                let maxItemWidth = legendDiscreteHelper.getMaxItemWidth(legData);
                constrainedMaxWidth = mathMin(maxItemWidth, maxWidth);
                chart.itemMaxWidth = constrainedMaxWidth;
                // #ZC1904 subtract the itemMargin
                chart.legendLabelMaxWidth = constrainedMaxWidth - colorBoxWidth - data.itemMargin * 2;

                // windows wraps unnecessary with floating point in dimension on horizontal legend
                // so we are adding 2px to the max width for avoiding unnecessary wrap
                if ($Browser.WINDOWS && isColumnView) {
                    let adjustMaxWidth = 2;
                    pushLegendInfo({ object: legendInfoObj.legend_grp, value: { maxWidth: maxWidth + adjustMaxWidth } });
                    legseries.style('max-width', `${maxWidth + adjustMaxWidth}px`);
                }
            }

            /**
             * *****************************************************************
             *                      Legend Expandable                          *
             * *****************************************************************
             */

            // get number of visible item count
            visibleItemCount = legendDiscreteHelper.getVisibleItemRenderCount(wrap, rowsCount, maxWidth, maxHeight, tempItemHeight);

            /* New Expandable
            let keepTitleAligned = title_data.keepAligned && isColumnView;
            let expandableSpec = {
                globalFont,
                bound,
                alignLegend,
                container,
                callbackArgs,
                padding: data.itemMargin,
                config: expandable,
                expanded: expandable._expanded,
                remainingCount: legData.length - visibleItemCount,
                isHorizontal: isColumnView,
                height: () => (keepTitleAligned ? mathMax(tempItemHeight, headerDimension.height) : tempItemHeight), // TODO: work
                target: () => legseries,
                margin: () => (keepTitleAligned && !expandable.keepAligned ? `0 0 0 ${headerWidth}` : 0),
                renderList: () => renderLegendList(legData),
                setLegendDimension: () => setLegendDimension(legseries.selectAll('li'))
            };
            let expandableObj = new Expandable(expandableSpec);
            expandableObj.create();
            expandableObj.draw(container);
            expandableObj.element.selectAll('.zc-expandable').style('font-size', `${utils.getVal(expandable.fontSize)}px`); // Actually this is an issue
            chart.expandableObj = expandableObj;


            if (visibleItemCount > 0 && !expandableObj.expanded()) {
                series = renderLegendList(legData.slice(0, visibleItemCount));
            } else {
                series = renderLegendList(legData);
            }

            var legendExpanded = expandableObj.expanded(),
            */

            let expandableUtils = expandableLegend(
                    chartObj,
                    wrap,
                    legData,
                    legseries,
                    legData_length,
                    maxWidth,
                    maxHeight,
                    isColumnView,
                    headerWidth,
                    renderLegendList,
                    setLegendDimension,
                    alignLegend,
                    floating,
                    container,
                    chartarea,
                    plotarea,
                    order
                ),
                remainingItemCount = mathMax(0, legData_length - visibleItemCount);

            if (legendExpandable && remainingItemCount > 0 && !chartObj.legend.allRendered) {
                series = renderLegendList(legData.slice(0, visibleItemCount));
            } else {
                series = renderLegendList(legData);
            }

            // #ZC1891, #ZC1952
            if (legendExpandable && remainingItemCount > 0) {
                expandableUtils.create(remainingItemCount);
            } else if ((legendExpandable && remainingItemCount === 0) || !legendExpandable) {
                expandableUtils.remove();
            }

            // *********************
            var legendExpanded = chartObj.legend.expanded,
                activeRows = 1;

            scrollWid = elementIsScrollable(legseries) || (legendExpandable && visibleItemCount < legData_length) ? scrollbarwid : 0;

            function setLegendDimension(series) {
                if (isColumnView) {
                    // horizontal align ------->

                    // reset the dimension for auto calculation, if any
                    legseries.styles({ width: null, height: null });

                    // set the max width for each item. it will solve text overlapping issue.
                    series.style('max-width', `${maxWidth}px`);
                    pushLegendInfo({ object: legendInfoObj.legend_grp.legend_nodes, value: { maxWidth, maxWidth } });

                    // set the calculated max width in the wrap element,
                    // so that when we set inline to the legseries[ul], it won't stress to the full width.
                    // will reset after calculating the inline content width
                    wrap.style('max-width', `${maxWidth}px`);

                    // calculate the actual content width
                    // previously we were calculating with each item and sum their width.

                    // apply dummy style to get the content width
                    // the border will set the max with to the parent
                    legseries.styles({ display: 'inline', 'border-top': '1px solid' });
                    let legendWidth = getBoundingRect(legseries).width;

                    legseries.styles({ display: titleAlignedStyles.display, 'border-top': null });

                    wrap.style('max-width', `${wrapMaxWidth}px`); // resetting it
                    pushLegendInfo({ object: legendInfoObj, value: { maxWidth: wrapMaxWidth } });

                    // do need the below 2 lines. to get rid of image diff setting the height.
                    activeRows = mathRound(legseries.node().offsetHeight / maxItemHeight) || 1;
                    legseries.style('height', `${activeRows * maxItemHeight + 2}px`);

                    // check the scroll width after setting the height; // phantomjs issue
                    scrollWid =
                        elementIsScrollable(legseries) || (legendExpandable && visibleItemCount < legData_length) ? scrollbarwid : 0;

                    // inline block needs and width. otherwise when it auto wraps, it sets the width to the max-width
                    legseries.style('width', `${legendWidth + scrollWid + 2}px`);

                    pushLegendInfo({
                        object: legendInfoObj.legend_grp,
                        value: {
                            preCalculationStyle: {
                                width: null,
                                height: null,
                                display: 'inline',
                                'border-top': '1px solid'
                            },
                            display: titleAlignedStyles.display,
                            'border-top': null,
                            height: activeRows * maxItemHeight + 2,
                            width: legendWidth + scrollWid + 2
                        }
                    });
                } else {
                    // reset the width first
                    legseries.style('width', null);

                    // calculate the max item width
                    let legendWidth = getBoundingRect(legseries.style('display', 'inline-block')).width;
                    legseries.style('display', titleAlignedStyles.display);
                    legseries.style('width', `${legendWidth + scrollWid + 2}px`);
                    legseries.style('height', null);

                    pushLegendInfo({
                        object: legendInfoObj.legend_grp,
                        value: {
                            preCalculationStyle: {
                                width: null,
                                display: 'inline-block'
                            },
                            display: titleAlignedStyles.display,
                            height: null,
                            width: legendWidth + scrollWid + 2
                        }
                    });
                }
            }
            pushLegendInfo({
                object: legendInfoObj,
                value: { maxWidth, maxWidth }
            });
            // set the legend dimension
            setLegendDimension(series);
            // then align the legend
            alignLegend(container, floating, maxWidth, maxHeight);

            // Add the overflow after calculating the dimension.
            // Because IE ads scroll bar on `overflow:auto` by default without checking max-[width,height].
            // So add overflow auto only if the scroll height is more than the offsetHeight [IE ONLY]

            var addOverFlowY = ((legendExpandable && legendExpanded) || !legendExpandable) === true;

            if ($Browser.IE && addOverFlowY) {
                addOverFlowY = legseries.node().scrollHeight - activeRows > legseries.node().offsetHeight;
            }

            legseries.style('overflow-x', 'hidden'); // this is not needed
            legseries.style('overflow-y', addOverFlowY ? 'auto' : 'hidden');
            pushLegendInfo({
                object: legendInfoObj.legend_grp,
                value: { addOverFlowY: addOverFlowY }
            });
        });
        return chart;
    }

    chart.dispatch = dispatch;

    chart.chartarea = function(_) {
        if (!arguments.length) {
            return chartarea;
        }
        chartarea = simpleClone(_);
        return chart;
    };

    chart.systemconf = function(_) {
        if (!arguments.length) {
            return systemconf;
        }
        systemconf = _;
        return chart;
    };

    chart.chartObj = function(_) {
        if (!arguments.length) {
            return chartObj;
        }
        chartObj = _;
        return chart;
    };

    chart.color = function(_) {
        if (!arguments.length) {
            return color;
        }
        color = _;
        return chart;
    };

    function alignLegend(container, floating = false, maxWidth, maxHeight) {
        let x = 0,
            y = 0;
        if (floating) {
            let legend = systemconf.legend;
            var dimension = chartObj.legendarea.element.node().getBoundingClientRect();
            x = mathMin(chartObj.canvasarea.width - dimension.width, legend.x);
            y = mathMin(chartObj.canvasarea.height - mathMin(dimension.height, maxHeight), legend.y);
        } else {
            let position = getLegendPosition(chartarea, chartObj);
            x = position[0];
            y = position[1];
        }
        container.styles({
            left: `${x}px`,
            top: `${y}px`
        });
    }

    function getVisibleItemCount(legseries) {
        var leg = legseries.node(),
            elements = leg.children,
            count = 0;

        for (var ii = 0, ll = elements.length; ii < ll; ii++) {
            // for overriding elementIsVisible (Zoho Reports)
            if (utils.elementIsVisible(elements[ii], leg)) {
                count++;
            } else {
                break;
            }
        }
        return count;
    }

    return chart;
};

export function pushLegendInfo(input) {
    var obj_val = input.value,
        obj_key = input.key;

    if (input.object) {
        if (typeof obj_val === 'object') {
            var obj;
            if (defined(obj_key)) {
                obj = input.object[obj_key] = input.object[obj_key] ? input.object[obj_key] : {};
            } else {
                obj = input.object;
            }
            Object.keys(obj_val).forEach(function(_k) {
                obj[_k] = obj_val[_k];
            });
        } else {
            obj = input.object;
            if (defined(obj_key)) {
                obj[obj_key] = obj_val;
            } else {
                obj = obj_val;
            }
        }
    }

    // obj = { ...obj, ...obj_val };
}

export var _discreteLegend = legend.discrete;
