// @zohocharts/zc/src/legend/legend-renderer-helpers.js
//
import { ORDINAL_DATATYPE } from '../core/datatype-constants';
import { NULL, NONE } from '../core/zc-constants';
import { DIV_ELEMENT, SVG_ELEMENT } from '../core/element-constants';
import { $ZCG } from '../core/zcglobal';
import { Registry } from '../core/zc-registry';
import { LegendRegistry } from './legend-registry';
import { noOperation } from '../core/zc-constants';
import { getLegendBound } from './legend-bound';
import { getFormattedValue } from '../formatter/formatter';
import { createElement, applyFont } from '../utils/zcsvgutils';
import { isHeatMapChartType } from '../events/events-utils';
// import { isLegendEnabled, getLegendConfig } from './legend-helper';
import { legend_helpers } from '../core/main-zc-init';
import { createColorAxis } from './color-legend/color-legend';
import { arrayFind, defined, invokeFunction, isObjectPropDefined } from '../utils/zcutils';
import { applyEffect, getChartEffect } from '../effects/zceffects';
import { Scale } from '../scale/scale-component';

export function collect(chartObj, spec) {
    let legendHighlight = Registry.getComponent('LegendHighlight'),
        legendConfig = chartObj.systemConf.legend,
        SUPPORTED =
            !chartObj.dataObject.hasZaxis || isHeatMapChartType(chartObj.chartTypes[0].value) || chartObj.dataObject.isFlowChartCategory
                ? []
                : ['z'],
        legendMeta = [],
        renderFunc = (domMode, id, data = [1]) => {
            const legendParent = spec && spec.parent || '';
            let isSvg = domMode === SVG_ELEMENT,
                eleType = isSvg ? 'g' : DIV_ELEMENT,
                isLastElement = (d, i, list) => i === list.length - 1, // Overview chart: on redraw 2 set of legends are drawn. legends are drawn for overview div too
                parent = chartObj.container.selectAll(`${legendParent} ${eleType}.legendarea`).filter(isLastElement);

            parent.style('display', NULL);
            return createElement(eleType, `.zc-legend-${id}`, data, parent);
        },
        renderTo = (domMode, id, data) => {
            let oppDomMode = domMode === SVG_ELEMENT ? 'html' : SVG_ELEMENT;
            renderFunc(oppDomMode, id, []); // remove opposite element
            let element = renderFunc(domMode, id, data);

            if (id.indexOf('z') > -1 || id.indexOf('shape') > -1 || (id.indexOf('clr') > -1 && domMode === SVG_ELEMENT)) {
                applyFont(element, [chartObj.systemConf.canvas, $ZCG], domMode !== SVG_ELEMENT);
            }

            return element;
        },
        dataSplicer = chartObj.dataSplicer,
        encodes = dataSplicer.getEncode();

    if (defined(chartObj.dataset.getColumnIdx('shape'))) {
        SUPPORTED.push('shape');
    }

    //Remove----
    createColorAxis(chartObj);
    var categoryHierarchy = chartObj.dataObject.categoryHierarchy,
        hasMultipleColorLegend,
        colorLegendLength = 1;

    if (categoryHierarchy) {
        var hierarchical = chartObj.dataObject.hierarchical,
            pivotLayout = hierarchical.pivotLayout,
            layers = pivotLayout.layers,
            colorLegendLength = layers.groups.length,
            hasMultipleColorLegend = colorLegendLength > 1;
    }
    let _ = 'clr',
        key = 0,
        key1 = 1;
    // if (encodes[_]) {
    //     encodes[_].forEach((index, key) => {
    // let id = `${_}-${key}`,
    //     id1 = `${_}-${key1}`,
    //     config = legendConfig;

    for (var i = 0; i < colorLegendLength; i++) {
        let id = `${_}-${i}`,
            config = legendConfig,
            view = _legendType(_, { config, dataType: chartObj.axes.clr.dataType }),
            domMode = LegendRegistry.get('clr', view).domMode,
            parser = isObjectPropDefined(chartObj.systemConf, `viewParser.legend.${id}`, true),
            column = chartObj.dataset.getColumn('clr'),
            legend_parser;

        if (i > 1 && legendConfig.color) {
            if (!legendConfig.color[0]) {
            }
            legendConfig.color[i] = config =
                legendConfig.color[i] != null ? mergeJSON(legendConfig.color[i], legendConfig.color[0], false) : legendConfig.color[0];
        }

        if(chartObj.viewFlow && chartObj.viewFlow.legends){
            for(let id in chartObj.viewFlow.legends){
                let legend_parser_conf = chartObj.viewFlow.legends[id].conf;
                if(column == undefined || column.columnname == legend_parser_conf.column){ //legend matched with parser conf
                    if(legend_parser_conf.column){
                        let legend_scale_id = legend_parser_conf.scale;

                        column = chartObj.viewFlow.getColumnByColumnName(legend_parser_conf.column);
                        //constructed legend scale via parser
                        if(legend_scale_id && !chartObj.viewFlow.scales[legend_scale_id].component){
                            var newScaleComp1 = new Scale(null, 0, chartObj.viewFlow.scales[legend_scale_id].conf, chartObj);
                            let newScaleObj = newScaleComp1.getScale();
                            chartObj.viewFlow.scales[legend_scale_id].component = newScaleObj;
                        }
                        legend_parser =  chartObj.viewFlow.legends[id];
                    }
                }
            }

            //assign legend parser if there is o scale assigned
            let availableLegendParserKeys = Object.keys(chartObj.viewFlow.legends);
            if(!legend_parser && availableLegendParserKeys.length == 1){
                legend_parser = chartObj.viewFlow.legends[availableLegendParserKeys[0]];
            }
        }

        

        let meta = {
            id,
            config,
            renderTo,
            parser,
            order: i,
            encode: _,
            globalFont: chartObj.systemConf.canvas,
            scaleComponent: chartObj.axes.clr,
            parserScaleComp: legend_parser && legend_parser.conf.scale ?  chartObj.viewFlow.scales[legend_parser.conf.scale].component : null,
            area: chartObj.chartarea,
            bound: getLegendBound(config, chartObj.chartarea),
            format: _format(column, chartObj, domMode),
            enabled: legend_helpers.isLegendEnabled(config, chartObj),
            view: legend_helpers.isContinousLegend(chartObj) ? 'band' : 'list',
            effect: getChartEffect(config, chartObj),
            callbackArgs: [chartObj],
            highlightManager: new legendHighlight(chartObj, chartObj.svg, { encode: 'clr', config, order: i }),
            // eventManager: legend.events(chartObj, element, states, disabledOpacity, i)
            chartObj, //remove
            legendView: legend_parser || (chartObj.viewFlow && chartObj.viewFlow.legends),
            column
        };
        legendMeta.push(meta);
    }
    //     });
    // }
    //----Remove

    SUPPORTED.forEach(_ => {
        if (encodes[_]) {
            encodes[_].forEach((index, key) => {
                let id = `${_}-${key}`,
                    scaleComponent = arrayFind(chartObj.scales, _ => _.id === id),
                    config = legend_helpers.getLegendConfig(_, key, chartObj),
                    view = _legendType(_, { config, dataType: scaleComponent.schema.data.dataType }),
                    domMode = LegendRegistry.get(_, view).domMode,
                    parser = isObjectPropDefined(chartObj.systemConf, `viewParser.legend.${id}`, true);

                let meta = {
                    id,
                    scaleComponent,
                    config,
                    renderTo,
                    view,
                    parser,
                    encode: _,
                    globalFont: chartObj.systemConf.canvas,
                    area: chartObj.chartarea,
                    bound: getLegendBound(config, chartObj.chartarea),
                    format: _format(scaleComponent.schema.data.meta, chartObj, domMode),
                    enabled: legend_helpers.isLegendEnabled(config, chartObj, _),
                    effect: getChartEffect(config, chartObj),
                    highlightManager: new legendHighlight(chartObj, chartObj.svg, { encode: _, config }),
                    callbackArgs: [chartObj]
                };

                if (!legendConfig.singleLayout) {
                    legendMeta.push(meta);
                }
            });
        }
    });

    chartObj.legend.isSingleLayout = legendConfig.singleLayout || legendMeta.length <= 1; //Remove //In rendering of legend from 2.27 it is ignored but not removed entirely
    legendMeta[0].bound.adjust = !chartObj.legend.isSingleLayout;

    return legendMeta;
}

export function render(meta) {
    let { encode, id, enabled, highlightManager, config, effect } = meta,
        renderer = LegendRegistry.get(encode, meta.view);

    if (renderer !== noOperation) {
        let component = new renderer(meta),
            element = meta.renderTo(renderer.domMode, id, [config]);

        element.style('display', enabled ? NULL : NONE);

        if (enabled) {
            clearDom(element, meta);
            var marks = component.create(),
                markExtender = meta.legendView && meta.legendView.conf && meta.legendView.conf.markExtender;
                if (markExtender) {
                    var extendedMark = invokeFunction(markExtender, marks, meta);
                    Object.assign(component.encode, extendedMark);
                }
            
            component.draw(element);

            // Filter effects
            applyEffect(element, { effect });
        }

        return { enabled, element, component, meta, id, highlightManager };
    }
}

export function _legendType(encode, options) {
    let { config, dataType } = options;

    if (config.type) {
        return config.type;
    }

    switch (encode) {
        case 'z':
        case 'size':
            return config.ranges || dataType === ORDINAL_DATATYPE ? 'list' : 'nested';
        case 'color':
        case 'clr':
            return config.ranges || dataType === ORDINAL_DATATYPE ? 'list' : 'band';
        case 'shape':
            return 'list';
    }
}
Registry.setComponent('legend.helpers.legendGroupType', _legendType);

function _format(meta, chartObj, domMode) {
    if (!meta) {
        return;
    }

    let dataType = meta.datatype,
        formatConf = meta[dataType],
        arraySeperator = '-',
        toDecode = domMode === SVG_ELEMENT;

    return d => getFormattedValue(formatConf, d, dataType, chartObj, 'legend', toDecode, arraySeperator);
}

function clearDom(element, meta) {
    let { view } = meta,
        prevView = element.attr('zc-view');

    if (prevView && view !== prevView) {
        element.selectAll('*').remove();
    }
    element.attr('zc-view', view);
}
