// @zohocharts/zc/src/legend/legend-discrete-helpers.js
import { applyFont, createElement } from '../utils/zcsvgutils';
import { $ZCG } from '../core/zcglobal';
import { getBoundingRect, defined, encodeHTML, pick, decodeHTML, getDepth } from '../utils/zcutils';
import { colorscale_helpers } from '../axis/colorscale-helpers';
import { getLegendShowAsText } from './legend-utils';
import { mathFloor, mathRound } from '../core/zc-constants';
import { hasColorColumnInNonAxis } from '../utils/colorutils';
import { $Browser } from '../core/browser-detect';

/**
 *
 * @param chart
 * @param legData
 * @param legSeries
 * @param legData_length
 * @param colorBoxSize
 * @param legendAlign
 * @param lineHeight
 * @param ImpFontProp
 * @return {{getLegendItemText: getLegendItemText, setItemParent(*): void}}
 */
export const helpers = (chart, legSeries, legData, legData_length, colorBoxSize, legendAlign, lineHeight, ImpFontProp) => {
    let systemConf = chart.systemConf,
        data = systemConf.legend,
        dataset = chart.dataset,
        colorBoxWidth = colorBoxSize[0],
        showAs = data.showAs,
        colIdx = chart.dataset.getColumnIdx(
            chart.isAxisCategory ||
                chart.isGeoMap ||
                chart.isNonAxisMultiSeriesCategory ||
                colorscale_helpers.isQuantileColorScale(chart) ||
                hasColorColumnInNonAxis(chart)
                ? 'clr'
                : 'x'
        ),
        isColumnView = legendAlign.layout === 'horizontal',
        itemWithIsNumber = typeof data.itemWidth === 'number';

    /**
     * For creating list
     * @param d
     * @param i
     * @return {string|*}
     */
    const methods = {
        getLegendItemText: (d, i) => {
            if (colorscale_helpers.isQuantileColorScale(chart)) {
                var colorBandOptions = data.colorBand,
                    userLabels = colorBandOptions.labels;

                if (userLabels && userLabels.length) {
                    return encodeHTML(userLabels[i]); //check with unorders ranges input
                }
            }
            const _colIdx = chart.dataObject.isFlowChartCategory ? chart.isLevelAxisCategory ? dataset.getColumnIdx('level', getDepth(d.level) - 1)
                : dataset.getColumnIdx(d._format || 'x', [0, d._axisOrder]) : colIdx;

            var seriesPos =
                    chart.isAxisCategory || chart.isNonAxisMultiSeriesCategory || chart.isGeoMap || hasColorColumnInNonAxis(chart) ? i : 0,
                showAsText = getLegendShowAsText(d, i, data, chart),
                nodeData = d.original || d.node || d;
            return showAs ? showAsText : chart.dataset.getDataPointValue(chart, nodeData, _colIdx, seriesPos, null, 'legend');
        },

        /**
         *
         * @param item
         * @return {{item: {width: *, height: *}, label: {width: number, height: number}}}
         */
        getItemDimension: item => {
            let { ul, li, span } = createTempItems([item]);

            let dim = {
                label: {
                    width: span.node().offsetWidth,
                    height: span.node().offsetHeight
                },
                item: {
                    width: li.node().offsetWidth,
                    height: li.node().offsetHeight
                }
            };
            ul.remove();
            // remove the temp element
            return dim;
        },

        /**
         *
         * @return {*}
         */
        getMaxItemWidth: () => {
            // we create only one element and put inner html in that and take dimension
            let tempElement = createTempItems([legData[0]]), // #ZC1911
                tempItemLabelElement = tempElement.span,
                itemElement = tempElement.li;

            let maxItemWidth = null,
                currentItemWidth;

            if (itemWithIsNumber) {
                return data.itemWidth;
            } else if (data.itemWidth === 'auto') {
                legData.forEach((d, i) => {
                    if (!d.hideInLegend) {
                        tempItemLabelElement.html(() => methods.getLegendItemText(d, i));

                        currentItemWidth = getBoundingRect(itemElement).width;

                        if (currentItemWidth > maxItemWidth) {
                            maxItemWidth = currentItemWidth;
                        }
                    }
                });
            }
            tempElement.li.remove();
            return maxItemWidth;
        },

        /**
         *
         * @param itemHeight
         * @param maxHeight
         * @return {number}
         */
        getRowsCount: (itemHeight, maxHeight) => {
            let actualRows = maxHeight / itemHeight,
                fullyPossibleRows = mathFloor(maxHeight / itemHeight) || 1,
                lastItemVisibility = actualRows - fullyPossibleRows,
                rowsCount = 0;

            // if the last item is visible more than 90%, then take that in account
            if (lastItemVisibility > 0.9) {
                rowsCount = mathRound(maxHeight / itemHeight) || 1;
            } else {
                rowsCount = mathFloor(maxHeight / itemHeight) || 1;
            }
            return rowsCount;
        },

        /**
         *
         * @param wrap
         * @param rowsCount
         * @param maxWidth
         * @param maxHeight
         * @param itemHeight
         * @return {number}
         */
        getVisibleItemRenderCount: (wrap, rowsCount, maxWidth, maxHeight, itemHeight) => {
            //
            let itemCount = 0;

            if (isColumnView) {
                let lastRowOverflow = false;

                if (itemWithIsNumber) {
                    let actualItemWidth = data.itemWidth + 2 * +data.itemMargin;
                    itemCount = mathFloor(maxWidth / actualItemWidth) * rowsCount; // plus 1 because we are rendering one extra element while auto calculate;
                } else {
                    let rowOverflow = false,
                        temp,
                        lastOffsetTop = null,
                        activeRow = -1;
                    while (!lastRowOverflow) {
                        itemCount++;

                        let legItems = legData.slice(0, itemCount);
                        temp = createTempItems(legItems, maxWidth, maxHeight);

                        let lastNode = temp.li.last().node(),
                            top = lastNode.offsetTop;

                        if (lastOffsetTop !== top) {
                            lastOffsetTop = top;
                            activeRow++;
                        }

                        rowOverflow = activeRow > rowsCount - 1;
                        lastRowOverflow = legData_length === itemCount || rowOverflow;
                    }
                    if (rowOverflow) {
                        itemCount -= 1;
                    }
                    temp.ul.remove();
                }
            } else {
                itemCount = rowsCount;
            }

            return itemCount;
        }
    };

    /**
     *
     * @param items
     * @param maxWidth
     * @param maxHeight
     * @return {{li: *, span: *}}
     */
    const createTempItems = (items, maxWidth, maxHeight) => {
        let ul = createElement('ul', '.tempUl', [1], chart.d3container).styles({
            'max-width': defined(maxWidth) ? `${maxWidth}px` : null, // ie9 issue
            'max-height': defined(maxHeight) ? `${maxHeight}px` : null, // ie9 issue
            position: 'absolute',
            // opacity: 0,
            top: 0,
            left: 0,
            padding: 0 // [#ZC2000] fix global padding conflict
        });

        let tempElement = createElement('li', '.tempItem', items, ul).styles({
            display: isColumnView ? 'inline-block' : 'block',
            'line-height': lineHeight,
            'white-space': 'nowrap',
            padding: `${data.itemMargin}px`,
            width: itemWithIsNumber ? `${data.itemWidth}px` : null,
            height: defined(data.itemHeight) ? `${data.itemHeight}px` : null,
            'padding-left': `${colorBoxWidth + data.itemMargin}px`
        });

        let span = createElement('span', '.tempLabel', d => [d], tempElement)
            .style('padding-left', `${data.textPadding}px`)
            .style('display', systemConf.legend.showAs ? 'block' : null);

        if (systemConf.legend.showAs) {
            var _styles = {
                // #ZC2002
                display: pick(systemConf.legend.showAsDisplay, 'inline-block'),
                overflow: 'hidden'
            };
            if(!$Browser.PHANTOMJS){
                _styles["vertical-align"] = "top";
            }
            span.styles(_styles);
        }

        // [#ZC2000] propagate data and index
        tempElement.select('span.tempLabel').html((d, i) => methods.getLegendItemText(d, i));

        // apply font
        applyFont(tempElement, [ImpFontProp, data, systemConf.canvas, $ZCG], true);

        return {
            ul,
            li: tempElement,
            span
        };
    };

    return methods;
};
