// @zohocharts/zc/src/legend/legend-brush-options.js
//$Id$
//legend-brush-options.js
import { mergeJSON } from '../utils/zcutils';
import { TRANSPARENT } from '../core/zc-constants';
import { getDefaultSliderOptions } from '../behaviours/visual-effects/brush-handles/brush-handle-options';
import { getLegendDefaultSliderOptionsArguments } from './legend-slider-arguments';
import { getLegendType } from '../axis/colorscale-helpers';

export function colorBandBrushOptions(config, bound, dataType, area) {
    var trackOptions = {
            fillColor: TRANSPARENT
        },
        sliderOptions = {
            type: 'lever',
            shadow: '0 0 2px rgba(0,0,0,0.2)',
            tooltip: {
                ...brushTooltipOptions(),
                enabled: getLegendType(config, dataType) === 'continuous'
            }
        };

    return {
        track: trackOptions,
        slider: mergeJSON(
            getDefaultSliderOptions.apply(null, getLegendDefaultSliderOptionsArguments(config, bound, area || dataType)), // TODO: change area, dataType
            sliderOptions,
            true
        )
    };
}

export function brushTooltipOptions() {
    return {
        enabled: true,
        padding: 5,
        margin: 5,
        backgroundColor: 'white',
        borderColor: 'rgb(51, 51, 51)',
        fontColor: 'rgb(51, 51, 51)'
    };
}
