// @zohocharts/zc/src/legend/legend-expandable.js
import { appendEle } from '../utils/zcsvgutils';
import { defined, getValByPriority, invokeFunction, utils, arrayFind } from '../utils/zcutils';
import { POINTER, PX } from '../core/zc-constants';
import { $ZCG } from '../core/zcglobal';
import { updateChartAndPlotBound } from './legend-bound';
import { axis, legend } from '../core/main-zc-init';
import { getLegendArea } from './legend-init';
import { pushLegendInfo } from './legend-discrete';

export function expandableLegend(
    chart,
    wrap,
    legData,
    legseries,
    legData_length,
    maxWidth,
    maxHeight,
    isColumnView,
    headerWidth,
    renderLegendList,
    setLegendDimension,
    alignLegend,
    floating,
    container,
    chartarea,
    plotarea,
    order
) {
    let systemConf = chart.systemConf,
        data = systemConf.legend,
        priorityArray = [data, systemConf.canvas, $ZCG],
        expandable = data.expandable || {},
        legendExpandable = expandable.show === true,
        title_data = data.title,
        keepLegendTitleAligned = title_data.keepAligned && isColumnView,
        redrawing = !!chart.renderer.length;

    return {
        /**
         *
         * @param remainingCount
         * @return {*}
         */
        create: remainingCount => {
            var legendInfoObj = chart.legend.unitTestObject['zc-legend'].clr[order];
            if (legendExpandable) {
                legseries.style('overflow', 'hidden');

                let showMoreElement = appendEle(wrap, 'div', [1], '.', 'class', 'expandable').styles({
                        color: expandable.fontColor,
                        'font-size': `${utils.getVal(expandable.fontSize)}px`,
                        'font-family': expandable.fontFamily || getValByPriority(priorityArray, 'fontFamily'),
                        'text-align': expandable.textAlign,
                        cursor: POINTER,
                        display: 'block',
                        padding: `${data.itemMargin}px`,
                        overflow: 'hidden',
                        'text-overflow': 'ellipsis',
                        'white-space': 'nowrap',
                        'max-width': `${maxWidth}px`,
                        'margin-left': keepLegendTitleAligned ? headerWidth + PX : 0
                    }),
                    legendExpanded = chart.legend.expanded;

                var content;
                if (!legendExpanded) {
                    content = invokeFunction(expandable.text, remainingCount, chart) || `${remainingCount} more...`;
                    showMoreElement.html(content);
                } else if (redrawing && legendExpanded) {
                    showMoreElement.html(function() {
                        content = invokeFunction(expandable.ontoggle, this, legendExpanded, chart) || 'Show less..';
                        return content;
                    });
                }

                pushLegendInfo({
                    object: legendInfoObj,
                    key: 'expandable',
                    value: {
                        maxWidth: maxWidth,
                        'margin-left': keepLegendTitleAligned ? headerWidth + PX : 0,
                        html: content
                    }
                });

                showMoreClickAction(remainingCount, showMoreElement);

                return showMoreElement;
            } else {
                if (!wrap.select('.expandable').empty()) {
                    wrap.selectAll('.expandable').remove();
                    legseries.style('overflow-y', 'hidden');
                    legseries.style('overflow-x', 'hidden');
                }
            }
        },
        remove: () => {
            if (!wrap.select('.expandable').empty()) {
                wrap.selectAll('.expandable').remove();
                legseries.style('overflow-y', 'hidden');
                legseries.style('overflow-x', 'hidden');
            }
        }
    };

    function updateChart() {
        if (chart.legend.isSingleLayout) {
            chart.updatePlotArea();
            updateChartAndPlotBound(chart, true);
        } else {
            chart.chartarea = chartarea;
            chart.plotarea = plotarea;
        }

        if (chart.isAxisCategory) {
            axis.render(chart, chart.chartWidth, chart.chartHeight, true);
        } else if (chart.dataObject.isPolarAxisCategory) {
            axis.polar.render(chart, 'x');
        }
        chart.drawThreshold(true);
        chart.drawPlotareaClippath();
        chart.updatePlotareaProp();
        chart.plot.renderer.draw(true);

        if (chart.isAxisCategory || chart.dataObject.isPolarAxisCategory) {
            chart.positionEventTrackerElement();
        }

        // chart.eventHandler.updateParams(chart);
    }

    function showMoreClickAction(remainingCount, showMoreElement) {
        showMoreElement.on('click', function() {
            if (!chart.legend.allRendered && remainingCount > 0) {
                //Reset cache for data
                chart.cache.data = {};
                chart.cache.search = {};
                chart.cache.correction = {};

                chart.legend.allRendered = true;

                let series = renderLegendList(legData);
                setLegendDimension(series);
                if (chart.legend.isSingleLayout) {
                    alignLegend(container);
                } else {
                    legend.align(chart.legends, chart, chartarea, plotarea);
                }

                if (!isColumnView && container.node().offsetWidth > chart.legendarea.width) {
                    let colorLegend = arrayFind(chart.legends, _ => _.id === 'clr-0');
                    chart.legendarea = getLegendArea(chart, colorLegend); // Adapting old method // Remove
                    updateChart();
                }
            }

            let legendExpanded = (chart.legend.expanded = !chart.legend.expanded);
            var updatedText = invokeFunction(expandable.ontoggle, this, legendExpanded, chart);
            if (defined(updatedText)) {
                updatedText = `${updatedText}`;
            }

            if (legendExpanded) {
                this.innerHTML = updatedText || 'Show less..';
                legseries.style('overflow-y', 'auto');
            } else {
                legseries.node().scrollTop = 0;
                this.innerHTML = updatedText || invokeFunction(expandable.text, remainingCount, chart) || `${remainingCount} more...`;

                legseries.style('overflow-y', 'hidden');
            }
        });
    }
}
