// @zohocharts/zc/src/legend/legend-helper.js
//$Id$
//legend-helper.js

import { scale, charttype } from '../core/main-zc-init';
import {
    defined,
    isArray,
    pInt,
    d3_values,
    perToPx,
    jsonStringify,
    pick,
    splat,
    arrayFind,
    objectValues,
    arrayIncludes,
    isObjectPropDefined,
    isObject
} from '../utils/zcutils';
import { ORDINAL_DATATYPE } from '../core/datatype-constants';
import { rgb2hex, getColorByIndex, hasColorColumnInNonAxis } from '../utils/colorutils';
import { mathMin, NONE, NULL, UNDEFINED } from '../core/zc-constants';
import { colorscale_helpers, getLegendType } from '../axis/colorscale-helpers';
import { isMultiPieChartType } from '../events/events-utils';
import { dimension_based_intelligence } from '../utils/intelligence-utils';
import {
    ZC_LEGEND_maxWidthHoriRatio,
    ZC_LEGEND_maxHeightHoriRatio,
    ZC_LEGEND_maxCenterWidth,
    ZC_LEGEND_maxCenterHeight,
    ZC_LEGEND_maxHeightVertRatio,
    ZC_LEGEND_maxWidthVertRatio
} from './legend-constants';
import { TOP_ALIGN, BOTTOM_ALIGN, LEFT_ALIGN, RIGHT_ALIGN, HORIZONTAL_ALIGN } from '../core/align-constants';
import { PIE_CHARTNAME, FUNNEL_CHARTNAME, PYRAMID_CHARTNAME } from '../core/chartname-constants';
import { isWithinEncode, isDataActive } from '../core/data-processor/is-data-active';
import { convertNull } from '../core/data-processor/size/convert-null';
import { getGridList } from '../core/data-processor/pivot-layout-helpers';
import { isFacet } from '../axis/hierarchy-axis-utils';
import { fadeOnSelectAll } from './helpers/select-all/fade-on-select-all';
import { legend_helpers } from '../core/main-zc-init';
import { getDefaultLegendMaxDimension, getLegendAlignment } from './legend-utils';

// export var legend_helpers = {};  //! Moved to main-zc-init
legend_helpers.getLegendData = function(chartObj, order) {
    // Continuous color - ranges
    if (colorscale_helpers.isQuantileColorScale(chartObj)) {
        var userRanges = chartObj.systemConf.legend.colorBand.ranges;
        if (userRanges && userRanges.length) {
            return userRanges;
        }
    }

    //Pivot - non axis chart
    let { dataObject } = chartObj;
    if (
        chartObj._isAxisCategory() &&
        !dataObject.isAxisCategory &&
        !dataObject.isPolarAxisCategory &&
        !dataObject.isNonAxisMultiSeriesCategory &&
        !hasColorColumnInNonAxis(chartObj)
    ) {
        let gridList = getGridList(dataObject);
        return gridList[order || 0].data[0].data[0]; //TODO: check for missing data in the first grid
    }

    var categoryHierarchy = chartObj.dataObject.categoryHierarchy;
    if (categoryHierarchy) {
        var hierarchical = chartObj.dataObject.hierarchical,
            pivotLayout = hierarchical.pivotLayout,
            rowList = pivotLayout.getRootNodeList('row'),
            layers = pivotLayout.layers,
            isLayerNode = layers.isLayerNode,
            hasMultiClusterRows = rowList.hasMultiCluster(),
            hasMultiLayers = layers.groups.length > 1,
            panes = layers.panes,
            gridlist = objectValues(panes);
        if (hasMultiLayers || (hasMultiClusterRows && isLayerNode)) {
            return gridlist[order || 0].data;
        }

        if (hasColorColumnInNonAxis(chartObj)) {
            return chartObj.seriesdata;
        }
    }
    if (chartObj.isAxisCategory || chartObj.isNonAxisMultiSeriesCategory || chartObj.isGeoMap) {
        return chartObj.seriesdata;
    } else if (chartObj.dataObject.isFlowChartCategory) {
        return chartObj.cache.sankey.legend;
    } else {
        return chartObj.seriesdata[0].data[0];
    }
};

legend_helpers.getLegendDataLength = function(chartObj, legData) {
    if (chartObj.isAxisCategory || chartObj.isNonAxisMultiSeriesCategory) {
        return legData.filter(function(d) {
            return !legend_helpers.isHiddenItem(d);
        }).length;
    } else {
        return legData.length;
    }
};

legend_helpers.isHiddenItem = function(d) {
    return d && d.hideInLegend;
};

legend_helpers.getClr = function(disabled, defaultClr, disabledOptions) {
    return disabled ? disabledOptions.color || defaultClr : defaultClr;
};

legend_helpers.updateBubblePieData = function(d, i, chartObj) {
    if (chartObj.dataObject.pieWithAxis) {
        d.data.forEach(function(subseries) {
            var isBubblePie = isMultiPieChartType(chartObj.chartTypes[i].value);
            subseries.forEach(function(dataPoint) {
                dataPoint.disabled = d.disabled || !isBubblePie || dataPoint.dummy;
            });
        });
    }
};

legend_helpers.updateHierarchyData = function(d, i, chartObj) {
    if (chartObj.dataObject.isHierarchy) {
        chartObj.systemConf.seriesdata.chartdata[i].disabled = d.disabled;
    }
};

legend_helpers.updateSankeyData = function (d, i, chartObj) {
    const shl = chartObj.dataObject.sankeyHierarchyLayout;
    if (shl) {
        d.disabled ? shl.filter('disableItems')(d) : shl.filter('enableItems')(d);
        shl.dependencies.updateSankeyData();
    }
}
legend_helpers.updateFilterStore = function (d, i, chart) {
    const { filterStore } = chart.dataObject;
    filterStore?.x(i, !d.disabled);
}

legend_helpers.toggleChartEvents = function(chartObj, enable) {
    let hasMultipleColorLegend = isObjectPropDefined(chartObj.dataObject, 'hierarchical.pivotLayout.layers.groups.length', true) > 1;

    if (isFacet(chartObj) && hasMultipleColorLegend) {
        chartObj.container.selectAll('rect.zc-fade-overlay').style('pointer-events', enable ? NONE : 'all');
        return;
    }

    var val = enable ? null : NONE,
        elements = [];
    if (!chartObj.svg) {
        elements.push(chartObj.container);
    } else {
        elements.push(chartObj.eventHandler.trackerElement);
        elements.push(chartObj.svg);
    }
    elements.forEach(function(d) {
        d.style('pointer-events', val);
    });
};

legend_helpers.fadeChart = function(enable, chartObj, opacity, forceEnable, order) {
    fadeOnSelectAll(enable, chartObj, opacity, forceEnable, order);
};

legend_helpers.isDisabled = function(d, allSelected, isFirstClickAfterDeselection) {
    return !allSelected && !isFirstClickAfterDeselection ? !allSelected : legend_helpers.isItemDisabled(d);
};

legend_helpers.isItemDisabled = function(d) {
    let dt = d.node || d;
    return dt.disabled || dt.filtered || d.disabled || d.filtered;
};

legend_helpers.isAllEnabled = function(legData, allSelected) {
    if (!allSelected) {
        return allSelected;
    }
    var disabledEle = legData.filter(legend_helpers.isItemDisabled);
    return !disabledEle.length;
};

legend_helpers.getFilterBoxSize = function(config, _default) {
    let size = pick(config.size, _default),
        width = pick(config.width, size),
        height = pick(config.height, size);

    return [width, height];
};

legend_helpers.updateData = function(d, i, chartObj) {
    if (colorscale_helpers.isQuantileColorScale(chartObj)) {
        chartObj.axes.clr.scale = scale.getScale(chartObj, chartObj.dataObject, null, null, null, chartObj.dataset, 'clr');
    } else if (!isFacet(chartObj)) {
        legend_helpers.updateBubblePieData(d, i, chartObj); //For bubblepie chart, it is mandatory to mark every data point to be disabled if the series is disabled
        legend_helpers.updateHierarchyData(d, i, chartObj); //For hierarchy chart, chartObj.seriesdata reference is not equal to chartObj.systemConf.seriesdata.chartdata reference as the chartObj.seriesdata is simplecloned. So, storing disable information at systemconf is mandatory
        legend_helpers.updateSankeyData(d, i, chartObj);
        legend_helpers.updateFilterStore(d, i, chartObj);
        chartObj.dataObject.data = chartObj.systemConf;
    }
};

legend_helpers.isContinousLegend = function(chartObj) {
    if (colorscale_helpers.isQuantileColorScale(chartObj)) {
        let config = chartObj.systemConf.legend,
            type = getLegendType(config, chartObj);

        return type === 'continuousDiscrete' || type === 'continuous';
    }

    return false;
};

legend_helpers.getColorBandDimension = function(config, bound, chartObj) {
    var chartarea = chartObj.chartarea || chartObj, // TODO: remove chartObj
        legendOptions = config,
        legendAlignment = getLegendAlignment(legendOptions),
        colorbandOptions = legendOptions.colorBand,
        isHorizontal = legendAlignment.layout === 'horizontal',
        { maxWidth, maxHeight } = bound,
        w = colorbandOptions.width,
        h = colorbandOptions.height;

    if (isHorizontal) {
        w = w || '40%';
        h = h || 14;
    } else {
        w = w || 14;
        h = h || '70%';
    }

    //Remove---
    // Adopting old
    if (!bound.adjust) {
        let legendMaxDimension = getDefaultLegendMaxDimension(legendOptions);
        maxWidth = legendMaxDimension.width;
        maxHeight = legendMaxDimension.height;
    }
    //---Remove

    maxWidth = perToPx(maxWidth, chartarea.width);
    maxHeight = perToPx(maxHeight, chartarea.height);

    if (bound.adjust) {
        let percent = 0.8;
        if (isHorizontal) {
            maxWidth *= percent;
        } else {
            maxHeight *= percent;
        }
    }

    w = perToPx(w, chartarea.width);
    h = perToPx(h, chartarea.height);

    w = mathMin(w, maxWidth);
    h = mathMin(h, maxHeight);

    return [w, h];
};

legend_helpers.getOrient = function(legendOptions) {
    var align = getLegendAlignment(legendOptions),
        hAlign = align.hAlign,
        vAlign = align.vAlign,
        floating = align.floating,
        isHorizontal = align.layout === HORIZONTAL_ALIGN,
        orient,
        columnViewOrients = {
            top: 'top',
            center: 'bottom',
            bottom: 'bottom'
        },
        rowViewOrients = {
            left: 'left',
            center: 'right',
            right: 'right'
        };

    if (floating) {
        orient = isHorizontal ? BOTTOM_ALIGN : RIGHT_ALIGN;
    } else if (isHorizontal) {
        // -->
        orient = columnViewOrients[vAlign];
    } else {
        orient = rowViewOrients[hAlign];
    }

    return orient;
};

legend_helpers.getLabelIndex = function(value, colorbandOptions, dataType) {
    var userStops = colorbandOptions.stops,
        userRanges = colorbandOptions.ranges,
        labelIndex;

    if (userRanges && userRanges.length) {
        userRanges.some(function(d, i) {
            var condition;
            if (isArray(d)) {
                if (isArray(value)) {
                    condition = d[0] === value[0] && d[1] === value[1];
                } else {
                    condition = d[0] === value;
                }
            } else {
                condition = d === value;
            }
            if (condition) {
                labelIndex = i;
                return true;
            }
        });
    } else if (dataType !== ORDINAL_DATATYPE && userStops && userStops.length) {
        userStops.some(function(d, i) {
            var condition = d === value;
            if (condition) {
                labelIndex = i;
                return true;
            }
        });
    }

    return labelIndex;
};

legend_helpers.hexToRGB = function(color) {
    //Converting hex colors to rgb
    if (color && color.toString().indexOf('#') == -1) {
        return color;
    }
    var bigint = pInt(rgb2hex(color).replace(/#/g, ''), 16);
    return `rgb(${(bigint >> 16) & 255},${(bigint >> 8) & 255},${bigint & 255})`;
};

legend_helpers.filterSelectionByColor = function(selection, selectedSeries, filterIndex, chartObj, encode = 'clr') {
    let options;

    if (encode === 'clr') {
        let config = chartObj.systemConf.legend;
        options = {
            encode,
            config,
            userRanges: config.colorBand.ranges,
            scale: chartObj.axes.clr.scale,
            dataType: chartObj.dataset.getDataType('clr'),
            dataAccessor: d => chartObj.dataset.getClr(d),
            valueByIdex: i => getColorByIndex(chartObj, i)
        };
    } else if (encode === 'z') {
        let scaleComp = arrayFind(chartObj.scales, _ => _.id === 'z-0'),
            scale = scaleComp.getScale(),
            scale_range = scale.range(),
            legendConfig = chartObj.systemConf.legend,
            config = legendConfig.size[scaleComp.key] || legendConfig.size;

        options = {
            encode,
            config,
            scale,
            userRanges: config.ranges,
            dataType: chartObj.dataset.getDataType('z'),
            dataAccessor: d => {
                let val = chartObj.dataset.getZ(d),
                    nullAsOptions = { config, scale, columnKey: chartObj.dataset.getColumnIdx(encode) };
                val = convertNull(val, nullAsOptions);
                return val;
            },
            valueByIdex: i => scale_range[i]
        };
    }

    return filterSelection(selection, selectedSeries, filterIndex, chartObj, options);
};

function filterSelection(selection, selectedSeries, filterIndex, chartObj, options) {
    if (defined(filterIndex)) {
        let { config, scale, dataType, dataAccessor, userRanges, encode, valueByIdex } = options,
            getValue = d => dataAccessor((d._acutalData && d._acutalData()) || d.data || d); // ZC2729 _actualData is only for Pyramid chart as data is reversed for Pyramid

        if (userRanges && userRanges.length) {
            var filterValue = splat(userRanges[filterIndex]),
                isOrdinal = dataType === ORDINAL_DATATYPE;

            if (filterValue.length === 1 && !isOrdinal) {
                var domain = d3_values(scale.domain());
                domain.splice(0, 2);
                filterValue = [domain[filterIndex * 2], domain[filterIndex * 2 + 1]];
            }

            return selection.filter(function(d) {
                var val = getValue(d),
                    condition = isDataActive(d, chartObj) && isWithinEncode(val, encode, chartObj, scale, config, { range: filterValue });

                if (isOrdinal) {
                    condition = condition && val === filterValue[0];
                } else {
                    condition = condition && val >= filterValue[0] && val <= filterValue[1];
                    if (val === filterValue[0]) {
                        condition = condition && valueByIdex(filterIndex) === scale(val);
                    }
                    if (val === filterValue[1]) {
                        condition = condition && valueByIdex(filterIndex) === scale(val);
                    }
                }
                return selectedSeries ? condition : !condition;
            });
        } else {
            return selection.filter(function(d) {
                var val = getValue(d),
                    condition = isDataActive(d, chartObj) && isWithinEncode(val, encode, chartObj, scale, config, { range: filterIndex });

                return selectedSeries ? condition : !condition;
            });
        }
    } else {
        return selection.filter(d => isDataActive(d, chartObj));
    }
}

legend_helpers.isReversedColorAxis = function(config) {
    return config.colorAxis && config.colorAxis.reversed;
};

legend_helpers.hasPatternPallete = function(chartObj) {
    var patternPallete = chartObj.patternPallete;
    return patternPallete && patternPallete.length;
};

// /**
//  *
//  * @param legendOptions
//  * @return {{width: number, height: number}}
//  */
// export function getDefaultLegendMaxDimension(legendOptions) {
//     var legendAlign = getLegendAlignment(legendOptions),
//         vAlign = legendAlign.vAlign,
//         hAlign = legendAlign.hAlign,
//         isHorizontalLayout = legendAlign.layout === 'horizontal',
//         widthRatio = 0,
//         heightRatio = 0;

//     if (isHorizontalLayout) {
//         if (vAlign === 'top' || vAlign === 'bottom') {
//             widthRatio = ZC_LEGEND_maxWidthHoriRatio;
//             heightRatio = ZC_LEGEND_maxHeightHoriRatio;
//         } else if (vAlign === 'center') {
//             widthRatio = ZC_LEGEND_maxCenterWidth;
//             heightRatio = ZC_LEGEND_maxHeightVertRatio;
//         }
//     } else {
//         if (hAlign === 'left' || hAlign === 'right') {
//             widthRatio = ZC_LEGEND_maxWidthVertRatio;
//             heightRatio = ZC_LEGEND_maxHeightVertRatio;
//         } else if (hAlign === 'center') {
//             widthRatio = ZC_LEGEND_maxWidthHoriRatio;
//             heightRatio = ZC_LEGEND_maxCenterHeight;
//         }
//     }

//     return {
//         width: widthRatio,
//         height: heightRatio
//     };
// }

// /**
//  *
//  * @param legendConfig
//  * @returns {{hAlign: *, vAlign: *, layout: *}}
//  */
// export function getLegendAlignment(config) {
//     // fix center top
//     var { hAlign, vAlign, layout: view, x, y } = config,
//         floating = defined(x) && defined(y);

//     // keep existing support
//     if (hAlign === 'center' && vAlign === 'center') {
//         hAlign = 'right';
//     }

//     if (floating) {
//         hAlign = LEFT_ALIGN;
//         vAlign = TOP_ALIGN;
//     }

//     if (!defined(view)) {
//         if (floating) {
//             view = 'vertical';
//         } else if (vAlign === 'top' || vAlign === 'bottom') {
//             view = 'horizontal';
//         } else {
//             view = 'vertical';
//         }
//     }

//     return {
//         hAlign: hAlign,
//         vAlign: vAlign,
//         layout: view,
//         floating: floating
//     };
// }

export function elementPriority(element, properties = []) {
    var className = 'zc-priority';
    if (arguments.length === 1) {
        return JSON.parse(element.attr(className)) || [];
    } else {
        element.attr(className, jsonStringify(properties));
    }
}

export function isLegendEnabled(config, chartObj, encoding = 'color') {
    var { vAlign, floating } = getLegendAlignment(config),
        dim = chartObj.canvasarea[vAlign === TOP_ALIGN || vAlign === BOTTOM_ALIGN ? 'height' : 'width'],
        intelligence = dimension_based_intelligence(chartObj, 'legend', dim);

    if ((floating || intelligence) && _encodingBased(encoding, chartObj)) {
        // #ZC2582
        if (chartObj.dataObject.isFlowChartCategory && config.enabled) {
            return chartObj.dataObject.sankeyLayout.nodeLength > 0;
        }

        return config.enabled;
    }

    return false;
}
legend_helpers.isLegendEnabled = isLegendEnabled;

export function hasAssociatedLegendData(chartObj) {
    let nonAxisSingleSeries =
        isFacet(chartObj) &&
        arrayIncludes([PIE_CHARTNAME, FUNNEL_CHARTNAME, PYRAMID_CHARTNAME], chartObj.chartTypes[0].name) &&
        !chartObj.userdata.facets.color.length;

    return isFacet(chartObj) && !nonAxisSingleSeries;
}

export function getLegendContainer(chartObj, id = 'clr-0') {
    if (isFacet(chartObj)) {
        let legendObj = chartObj.legends.get(id);
        return (legendObj && legendObj.element) || chartObj.legendarea.element;
    }

    return chartObj.legendarea.element;
}

export function getLegendConfig(encodeType, key, chart) {
    let encode = encodeType === 'z' ? 'size' : encodeType === 'clr' ? 'color' : encodeType,
        baseConfig = chart.systemConf.legend;

    return baseConfig[encode][key] || baseConfig[encode];
}
legend_helpers.getLegendConfig = getLegendConfig;

function _encodingBased(encoding, chartObj) {
    switch (encoding) {
        case 'color': {
            if (colorscale_helpers.isQuantileColorScale(chartObj)) {
                return true;
            }

            let multiSeries = chartObj.isAxisCategory || chartObj.isNonAxisMultiSeriesCategory || chartObj.isGeoMap, // #ZC1859
                accessor = d => !legend_helpers.isHiddenItem(d) && d.seriesname !== UNDEFINED,
                hasPivotLayerNode = true;

            if (chartObj.dataObject.pivot) {
                var layers = chartObj.dataObject.hierarchical.pivotLayout.getRootNodeList('layer');
                hasPivotLayerNode = !layers.noLayers;
            }
            return multiSeries
                ? chartObj.seriesdata.some(accessor) && hasPivotLayerNode
                : chartObj.chartTypes[0].value === charttype.dial
                ? chartObj.seriesdata[0].data[0].length > 1 && !isFacet(chartObj)
                : true;
        }
        default:
            return true;
    }
}
