// @zohocharts/zc/src/legend/legend-selectallbox.js
//$Id$
//legend-selectallbox.js
import { legend, legend_helpers } from '../core/main-zc-init';
import { hasAssociatedLegendData, getLegendContainer } from './legend-helper';
import { TRANSPARENT, POINTER, NONE } from '../core/zc-constants';
import { mergeJSON, getValByPriority, pick, invokeFunction, shorthand_px, defined, utils } from '../utils/zcutils';
import { appendEle } from '../utils/zcsvgutils';
import { getTooltipPosition, getEventObject } from '../events/events-utils';
import { getMousePosition } from '../events/eventActions';
import { $ZCG } from '../core/zcglobal';
import { getLegendShowAsText, legutil } from './legend-utils';
import { d3_event } from '../core/d3-constants';
import { getDisabledOptions } from './events/legend-event-helpers';
import { getGridList } from '../core/data-processor/pivot-layout-helpers';
import { getHTML4SVGSymbol, getSVGSymbolString } from '../utils/svg-symbols';

var getDefaultSelectAllOptions = () => ({
    show: false,
    strokeColor: '#262626',
    strokeWidth: 1,
    fillColor: TRANSPARENT,
    fontColor: '#262626',
    fillOpacity: 1,
    strokeOpacity: 1,
    disabledOpacity: 0.3,
    tooltip: {
        enabled: true,
        color: 'red',
        content: 'Select atleast one legend'
    },
    events: {
        click: null,
        tap: null
    }
});

var getLegendData = legend_helpers.getLegendData,
    isDisabled = legend_helpers.isDisabled,
    isAllEnabled = legend_helpers.isAllEnabled,
    getClr = legend_helpers.getClr,
    toggleChartEvents = legend_helpers.toggleChartEvents,
    updateData = legend_helpers.updateData;

legend.selectAllBox = function(container, chartObj, states, order) {
    // @INFO Memory Leak Fix
    var defaultSelectAllOptions = getDefaultSelectAllOptions();
    defaultSelectAllOptions.events.click = defaultSelectAllOptions.events.tap  /*= chartObj.legend.component.toggleSelectAll*/ = toggleSelectAll;

    if(chartObj.legend.component){
        chartObj.legend.component.toggleSelectAll  = toggleSelectAll;
    }
    var legData = getLegendData(chartObj, order),
        legendOptions = chartObj.systemConf.legend,
        colorBox = legendOptions.colorBox,
        selectAllBoxOptions = mergeJSON(defaultSelectAllOptions, legendOptions.selectAllBox, true),
        disabledOptions = getDisabledOptions(legendOptions, 'clr'),
        disabledColorBox = disabledOptions.colorBox || {},
        priorityArray = [legendOptions, chartObj.systemConf.canvas, $ZCG],
        fontSize = +utils.getVal(legendOptions.fontSize || chartObj.systemConf.canvas.fontSize),
        fontColor = getValByPriority(priorityArray, 'fontColor'),
        imagePallete = legendOptions.imagePallete ? legutil.imagePallete.get(legendOptions.imagePallete) : null,
        selectAllColorBoxMerged = mergeJSON(colorBox, selectAllBoxOptions, true),
        colorBoxSize = legend_helpers.getFilterBoxSize(selectAllColorBoxMerged, fontSize),
        colorBoxWidth = colorBoxSize[0],
        colorBoxHeight = colorBoxSize[1],
        selectAllFilterSize =
            Math.min(colorBoxWidth, colorBoxHeight) -
            (legendOptions.imagePallete || (!defined(legendOptions.shapes) || colorBox.shape === 'square') ? 0 : +colorBox.strokeWidth),
        svgNode = chartObj.container.node(),
        selectAllBoxTooltipOptions = selectAllBoxOptions.tooltip,
        disabledOpacity = selectAllBoxOptions.disabledOpacity,
        defaultTooltipColor;

    states.allSelected = defined(selectAllBoxOptions.selected) ? selectAllBoxOptions.selected : states.allSelected;

    var selectAllFilter = container && appendEle(container, 'div', [1], '.', 'class', 'selectAll');

    var shapeType = pick(selectAllBoxOptions.shape, colorBox.shape);
    var useSVGImage = shapeType !== 'square',
        styles = {
            display: selectAllBoxOptions.show ? 'inline-block' : NONE,
            width: `${selectAllFilterSize}px`,
            height: `${selectAllFilterSize}px`,
            float: 'left',
            background: selectAllBoxOptions.fillColor,
            position: 'absolute',
            'box-sizing': 'border-box',
            'text-align': 'center',
            'line-height': `${selectAllFilterSize}px`,
            'font-family': colorBox.fontFamily,
            'font-size': `${colorBox.fontSize || fontSize - 2}px`,
            color: function(d, i) {
                var _color =
                    selectAllBoxOptions.fontColor ||
                    (isDisabled(d, states.allSelected, states.isFirstClickAfterDeselection)
                        ? disabledColorBox.fontColor
                        : colorBox.fontColor);
                return _color;
            }
        };

    if (useSVGImage) {
        var path = getHTML4SVGSymbol(
            {
                ...selectAllBoxOptions,
                width: selectAllFilterSize,
                height: selectAllFilterSize,
                strokeColor: selectAllBoxOptions.strokeColor,
                fillColor: selectAllBoxOptions.fillColor,
                shape: shapeType
            }
        );
        styles.background = getSVGSymbolString(path, selectAllFilterSize, selectAllFilterSize);
        styles.border = 'none'; // #ZC1526
    } else {
        styles.border = `${selectAllBoxOptions.strokeWidth}px solid ${selectAllBoxOptions.strokeColor}`;
    }

    selectAllFilter &&
        selectAllFilter.styles(styles).html(function(d, i) {
            var content = isAllEnabled(legData, states.allSelected)
                ? pick(selectAllBoxOptions.text, colorBox.text) || '&#10003;'
                : disabledColorBox.text || '';
            return content;
        });

        container && container.style('cursor', selectAllBoxOptions.show ? POINTER : null);

    // var mouseHandlers = {
    //         click: selectBoxClickFunction
    //     },
    //     touchHandlers = {
    //         tap: selectBoxClickFunction
    //     };
    // if (selectAllBoxOptions.show) {
    //     attachevents(container, mouseHandlers, touchHandlers);
    // } else {
    //     detachevent(container, null, mouseHandlers);
    //     detachevent(container, null, touchHandlers);
    // }

    // if (chartObj.eventHandler) {
    //     var tooltip = chartObj.eventHandler.tooltipElement;
    //     tooltip.hide(tooltip.element);
    // }

    return {
        styles: styles,
        html: function() {
            var content = isAllEnabled(legData, states.allSelected)
                ? pick(selectAllBoxOptions.text, colorBox.text) || '&#10003;'
                : disabledColorBox.text || '';
            return content;
        },
        element: selectAllFilter,
        options: selectAllBoxOptions,
        size: selectAllFilterSize
    };

    function selectBoxClickFunction(selectAllFilter) {
        if (!(states.allSelected && !isAllEnabled(legData, states.allSelected))) {
            states.allSelected = !states.allSelected;
        }
        states.isFirstClickAfterDeselection = false;
        
        // selectAllFilter.html(
        //     states.allSelected ? pick(selectAllBoxOptions.text, colorBox.text) || '&#10003;' : disabledColorBox.text || ''
        // );
        var userFunction = selectAllBoxOptions.events[chartObj.zcEventType || 'click'];
        invokeFunction(userFunction, d3_event(), states.allSelected, chartObj);
    }

    function toggleSelectAll() {
        filterAllAxesNUpdate();
        applyStyleToFilterAndLabel();
        toggleChartEvents(chartObj, states.allSelected);
        legend_helpers.fadeChart(states.allSelected, chartObj, disabledOpacity);

        if (selectAllBoxTooltipOptions.enabled) {
            var tooltip = chartObj.eventHandler.tooltipElement;
            if (!defaultTooltipColor) {
                defaultTooltipColor = tooltip.element.selectAll('div#tooltipcontent').style('color');
            }
            var tooltip_data = chartObj.systemConf.tooltip;
            var tooltipBorderWidth = shorthand_px(defined(tooltip_data.borderWidth) ? tooltip_data.borderWidth : 1);
            if (!states.allSelected) {
                var coordinate = getMousePosition(svgNode, d3_event());
                coordinate = {
                    x: coordinate[0],
                    y: coordinate[1]
                };
                var evt = getEventObject(d3_event());
                var position = getTooltipPosition(evt, coordinate, 0, 0, chartObj, {}, false, false, false, true, false);
                var dLeft = coordinate.x - evt.clientX,
                    dTop = coordinate.y - evt.clientY;
                position.dLeft = dLeft;
                position.dTop = dTop;
                position.helpers.color = selectAllBoxTooltipOptions.color;
                position.helpers.backgroundColor = selectAllBoxTooltipOptions.backgroundColor;
                position.helpers.borderColor = selectAllBoxTooltipOptions.borderColor;
                position.helpers.renderTo = 'legendTitle';
                applyTooltipStyle(tooltip, 'center', null, selectAllBoxTooltipOptions.fontColor || defaultTooltipColor);
                tooltip.show(chartObj, position, selectAllBoxTooltipOptions.content, false, position.helpers);
                setTimeout(function() {
                    var tooltip = chartObj.eventHandler.tooltipElement;
                    applyTooltipStyle(tooltip, null, tooltipBorderWidth, defaultTooltipColor);
                    tooltip.hide(tooltip.element);
                }, 4000);
            } else {
                var tooltip = chartObj.eventHandler.tooltipElement;
                applyTooltipStyle(tooltip, null, tooltipBorderWidth, defaultTooltipColor);
                tooltip.hide(tooltip.element);
            }
        }
    }

    function filterAllAxesNUpdate() {
        if (!legData) {
            //If legend is disabled, legData will not be available
            legData = getLegendData(chartObj, order);
        }
        if (states.allSelected) {
            if (hasAssociatedLegendData(chartObj)) {
                var gridList = getGridList(chartObj.dataObject);

                gridList.forEach((grid, _gridIndex) => {
                    let { layer } = grid;

                    if (layer === order) {
                        grid.data.forEach((_d, _i) => {
                            _d.node.filtered = !states.allSelected;
                        });
                    }
                });

                legData.forEach((d, j) => (d.filtered = !states.allSelected));
            } else {
                legData.forEach(function(d, j) {
                    d.disabled = !states.allSelected;
                    updateData(d, j, chartObj);
                });
            }

            chartObj.addRemoveSeries(legData, 0, [0, 0], true);
        }
    }

    function applyStyleToFilterAndLabel() {
        var legendContainer = getLegendContainer(chartObj, `clr-${order}`);

        // #ZC1943
        let legendFilter = legendContainer.selectAll('div.legendfilterbox, div.zc-legend-legendfilterbox'),
            legendlabel = legendContainer.selectAll('span.legendlabel, span.legendlabel * , span.zc-legend-legendlabel, span.zc-legend-legendlabel * ');

        legend.setColorBoxBackground(legendFilter, chartObj, states, order);
        var fontClr, txtDec;
        if (states.allSelected) {
            (fontClr = fontColor), (txtDec = null);
        } else {
            (fontClr = getClr(true, fontColor, disabledOptions)), (txtDec = JSON.parse(disabledOptions.strikeout) ? 'line-through' : null);
        }
        legendlabel.styles({
            color: fontClr,
            'text-decoration': txtDec
        });

        // #ZC1914
        // for custom legend, while enabling legend all items, get the source from the user.
        // because user might set some different colors and can have different hierarchy
        if (states.allSelected && legendOptions.showAs) {
            legendContainer.selectAll('span.legendlabel, span.zc-legend-legendlabel').html((d, i) => getLegendShowAsText(d, i, legendOptions, chartObj));
        }
    }

    function applyTooltipStyle(tooltip, align, borderWidth, contentColor) {
        tooltip.element.attr('align', align).styles({
            'border-width': borderWidth
        });
        tooltip.element.selectAll('div#tooltipcontent').style('color', contentColor);
    }
};

export let _legendSelectAll = legend.selectAllBox;
