// @zohocharts/zc/src/scale/scale-component.js
//$Id$
//scale-component.js
import { NUMERIC_DATATYPE, ORDINAL_DATATYPE, TIME_DATATYPE } from '../core/datatype-constants';
import { NULL } from '../core/zc-constants';
import { DataProcessor } from '../core/data-processor';
import {
    d3_scaleLinear,
    d3_scaleOrdinalOrg,
    d3_scaleTime,
    d3_merge,
    d3_ascending,
    d3_extent,
    d3_scaleSqrt,
    d3_scaleOrdinal,
    d3_scaleSymlog
} from '../core/d3-constants';
import { extent } from './domain/extent';
import { categories } from './domain/categories';
import { colorscale_helpers } from '../axis/colorscale-helpers';
import { d3_values, defined, invokeFunction, simpleClone, isArray } from '../utils/zcutils';
import { generateDateVal } from '../formatter/formatter';
import { SQRT, LOG } from '../core/scale-type-constant';
import { QUANTILE_SCALETYPE } from './scale-type-constants';
import { isFunction } from '../utils/function-utils';
import { ALLOWED_OPERATORS } from './operators/allowed-operators';
import { ScaleRegistry } from './scale-registry';
import { isExpressionType, getExpressionValue } from '../mark/parser/parser-helpers';

export class Scale {
    constructor(encode, key, config, chartObj) {
        this.encode = encode || config.id || 'custom';
        this.key = key || 0;
        this.config = config;
        this.chartObj = chartObj;
        this.bounds = {};
        this.init();
    }

    init() {
        this.scale = NULL;
        this.id = `${this.encode}-${this.key}`;
        this.d3_scale = {};

        let config = this.getConfig(),
            data = this.chartObj && this.getDataSchema(),
            { ranges } = config;

        if (ranges && !isArray(ranges[0])) {
            config.ranges = ranges.map(r => [r]);
        }
        this.datatype = data && data.dataType || this.getDatatype();
        this.schema = {
            data: data,
            domain: {
                key: data && data.meta.dataindex,
                min: config.minRange,
                max: config.maxRange,
                ranges: config.ranges,
                stops: config.stops,
                base: config.base
            }
        };
    }

    create() {
        let scale = this.scale = this.getScaleType()(),
            config = this.getConfig();

        this.d3_scale.domain = this._parseDomain();
        this.d3_scale.range = this._parseRange();

        // this.d3_scale.unknown = 0; // ZC2428

        if (config.clamp) {
            this.d3_scale.clamp = true;
        }

        for (var _ in this.d3_scale) {
            if (scale[_]) {
                scale[_](this.d3_scale[_]);
            } else {
                scale[_] = this.d3_scale[_];
            }
        }
        this.scale.unknown?.(this.d3_scale.unknown = this._parseUnknown());

        return scale;
    }

    getScale() {
        if (!this.scale) {
            this.create();
        }

        return this.scale;
    }

    getDatatype() {
        let datatype = this.config.type || this.config.datatype,
            column = this.config.column;

            if(column && !datatype){
                datatype = this.chartObj.dataset.columnMappedByName[column].datatype || ORDINAL_DATATYPE;
                this.config.type = datatype;
                this.config.dataType = datatype;
            }
            return datatype;
    }

    getScaleType() {
        let schema = this.schema && this.schema.data,
            dataType = this.config.dataType || schema.dataType,
            scaleType = this.config.scaleType || (schema && schema.meta.scaleType);
        return _getScaleType(dataType, this, scaleType);
    }

    getConfig() {
        return this.config;
    }
    _parseDomain() {
        let _scale = this,
            _config = _scale.config,
            _field = _config.field,
            _domainConfig = _config.domain,
            _datatype = _config.type;

        if(!_domainConfig){
            return _defaultDomain(this.config.column, this, this.chartObj)
        }

        if (isArray(_domainConfig)) {
            return _domainConfig;
        }

        if (isFunction(_domainConfig)) {
            invokeFunction(_domainConfig, this);
            return;
        }

        if(isExpressionType(_domainConfig)){
            return getExpressionValue( null, [null, null], _domainConfig, null, this.chartObj.viewFlow);
        }

        let appliedOperators = {};
        Object.keys(_domainConfig.operators).map(operatorKey => {
            if (ALLOWED_OPERATORS.has(operatorKey)) {
                appliedOperators[operatorKey] = ALLOWED_OPERATORS.get(operatorKey).call(
                    this,
                    _domainConfig.operators[operatorKey],
                    appliedOperators
                );
            }
        });
        var evaluator = typeof _domainConfig.evaluator === 'string' ? invokeFunction(_domainConfig.evaluator, _scale): _domainConfig.evaluator;
        return evaluator.call(this, appliedOperators);
    }
    _parseRange() {
        let _scale = this,
            _config = _scale.config,
            _field = _config.field,
            _rangeConfig = _config.range,
            _datatype = _config.type;

        if (isArray(_rangeConfig)) {
            return _rangeConfig;
        }

        if (isFunction(_rangeConfig)) {
            invokeFunction(_rangeConfig, this);
            return;
        }

        if(isExpressionType(_rangeConfig)){
            return getExpressionValue( null, [null, null], _rangeConfig, null, this.chartObj.viewFlow);
        }

        let appliedOperators = {};
        ALLOWED_OPERATORS.keys().map(operator => {
            if (_rangeConfig.operators[operator]) {
                appliedOperators[operator] = ALLOWED_OPERATORS.get(operator).call(this, _rangeConfig.operators[operator], appliedOperators);
            }
        });
        var evaluator = typeof _rangeConfig.evaluator === 'string' ? invokeFunction(_rangeConfig.evaluator, _scale): _rangeConfig.evaluator;
        return evaluator.call(this, appliedOperators);
    }
    _parseUnknown() {
        const config = this.config || {};
        return ('unknown' in config) ? config.unknown : (this.encode !== 'z' ? this.scale(0) : 0);
    }
    //! Seems like dead-code. Need to check
    parseDomain() {
        let dataOptions = this.schema.data,
            source = invokeFunction(dataOptions.source),
            dataType = dataOptions.dataType,
            schema = this.schema.domain;

        if (dataType === TIME_DATATYPE) {
            schema = toTimestamp(schema, dataOptions.meta);
        }

        let _min = schema.min,
            _max = schema.max,
            _stops = schema.stops,
            _ranges = schema.ranges,
            _base = schema.base;

        if (_ranges) {
            return d3_merge(_ranges);
        }

        if (dataOptions.condition) {
            source = source.filter(dataOptions.condition);
        }

        switch (dataType) {
            case ORDINAL_DATATYPE: {
                return categories(source, this.schema.domain.key);
            }
            case TIME_DATATYPE:
            case NUMERIC_DATATYPE: {
                let rawExtent = extent(source, this.schema.domain.key),
                    domain = rawExtent;

                if (defined(_base)) {
                    domain = d3_extent([...domain, _base]);
                }

                if (defined(_min)) {
                    domain[0] = _min;
                }

                if (defined(_max)) {
                    domain[domain.length - 1] = _max;
                }

                if (_stops) {
                    var copy_userStops = d3_values(_stops);
                    copy_userStops.sort(d3_ascending);
                    var filteredStops = copy_userStops.filter(function(d) {
                        var condition = true;
                        if (defined(_min)) {
                            condition = condition && d >= _min;
                        }
                        if (defined(_max)) {
                            condition = condition && d <= _max;
                        }
                        return condition;
                    });

                    if (domain[0] < filteredStops[0] || !filteredStops.length) {
                        filteredStops.splice(0, 0, domain[0]);
                    }
                    if (domain[1] > filteredStops[filteredStops.length - 1]) {
                        filteredStops.splice(filteredStops.length, 0, domain[1]);
                    }
                    return filteredStops;
                }

                return domain;
            }
        }
    }

    //! Seems like dead-code. Need to check
    parseRange() {}

    getDataSchema() {
        let isCatNull = DataProcessor.helpers.isCatNull;
        let encode = this.encode,
            key = this.key,
            chartObj = this.chartObj,
            processor = chartObj.dataObject,
            parser = chartObj.dataset,
            isCatNullX = isCatNull(parser, 'x'),
            isCatNullY = isCatNull(parser, 'y', 0),
            meta = parser.getColumn(encode, key, this.config.column),
            zEncode = processor.encodes && processor.getEncode(encode, key), //need encodes support for hierarchy processor DEMO_PACKEDBUBBLE_7
            dataType = encode ? chartObj.dataset.getDataType(encode, key) : _dataType(meta);

        return {
            source: zEncode ? d3_merge(zEncode.seriesData.map(_ => d3_merge(_.data))) : [],
            condition: d => (hasGeo(chartObj) ? true : !(isCatNullX(parser.getActualX(d, key)) || isCatNullY(parser.getY(d, key)))),
            meta: simpleClone(meta),
            dataType
        };
    }
}

function _getScaleType(dataType, scaleIns, scaleType) {
    let _ranges = scaleIns.schema.domain.ranges,
        _encode = scaleIns.encode;

    switch (dataType) {
        case NUMERIC_DATATYPE: {
            if (_ranges || scaleIns.config.scaleType == QUANTILE_SCALETYPE) {
                return colorscale_helpers.quantileScale;
            }

            if (scaleType === SQRT) {
                return d3_scaleSqrt;
            }
            if (scaleType === LOG) {
                return d3_scaleSymlog;
            } else {
                return d3_scaleLinear;
            }
        }
        case ORDINAL_DATATYPE:
            if (scaleIns.config.scaleType === '_ordinal') {
                return d3_scaleOrdinalOrg;
            }
            if (scaleIns.config.scaleType === 'band') {
                return d3_scaleOrdinal;
            }

            return _encode == 'x' || _encode == 'y' ? d3_scaleOrdinal : d3_scaleOrdinalOrg;
        case TIME_DATATYPE:
            if (_ranges || scaleIns.config.scaleType == QUANTILE_SCALETYPE) {
                return colorscale_helpers.quantileScale;
            }

            return d3_scaleTime;
    }
}

function _dataType(meta) {
    let { useDiscrete, datatype } = meta;

    if (useDiscrete) {
        return ORDINAL_DATATYPE;
    }

    return datatype || ORDINAL_DATATYPE;
}

export function toTimestamp(schema, meta) {
    let dataType = _dataType(meta),
        formatConfig = meta[dataType] || {},
        inputFormat = formatConfig.inputformat,
        regexTimeStamp = /timestamp/i,
        convert = _ => generateDateVal(_, null, formatConfig, true).getTime();

    if (!(inputFormat && !regexTimeStamp.test(inputFormat))) {
        return schema;
    }

    let _min = schema.min,
        _max = schema.max,
        _stops = schema.stops,
        _ranges = schema.ranges;

    if (_ranges) {
        _ranges = _ranges.map(range => range.map(convert));
    }

    if (_stops) {
        _stops = _stops.map(convert);
    }

    if (defined(_min)) {
        _min = convert(_min);
    }

    if (defined(_max)) {
        _max = convert(_max);
    }

    return { ...schema, ranges: _ranges, stops: _stops, min: _min, max: _max };
}

function hasGeo(chartObj) {
    // TODO  change chartObj.dataObject.isFlowChartCategory
    return chartObj.isGeoMap || chartObj.mapObj || chartObj.dataObject.isFlowChartCategory;
}

function _defaultDomain(columnName, scaleObj, chartObj){
    let domain,
        column = chartObj.dataset.columnMappedByName[columnName];
    if(column.datatype == NUMERIC_DATATYPE || column.datatype == TIME_DATATYPE){
        domain = ALLOWED_OPERATORS.get('extremes').call(
            scaleObj,
            {
                "extremes": [{
                    "_type": "extremesByData",
                    "indices": [column.dataindex]
                }]
            }
        );
    }else if(column.datatype == ORDINAL_DATATYPE){
        domain = ALLOWED_OPERATORS.get('categories').call(
            scaleObj,
            {
                "indices": [column.dataindex]
            }
        );
    }
    return domain;

}
ScaleRegistry.add('Scale', Scale);
