// @zohocharts/zc/src/mark/parser/parser-helpers.js
//
import { d3_map } from '../../core/d3-constants';
import { NULL, TRANSPARENT } from '../../core/zc-constants';
import { defined, getVal, isArray, isObject, pick } from '../../utils/zcutils';

export const _defaultMarkValue = {
    x: 0,
    y: 0,
    width: 0,
    height: 0,
    angle: 0,
    radius: 0,
    size: 0,
    points: [],
    fillColor: TRANSPARENT,
    strokeColor: TRANSPARENT
};

function cssToJsObject(cssObj) {
    let jsObj = {};
    for (var attr in cssObj) {
        let slashPos = attr.indexOf('-');
        if (slashPos > -1) {
            jsObj[attr.slice(0, slashPos) + attr[slashPos + 1].toUpperCase() + attr.slice(slashPos + 2)] = cssObj[attr];
        } else {
            jsObj[attr] = cssObj[attr];
        }
    }
    return jsObj;
}

// props => radius | [radius, borderRadius]
function getMarkValue(node, args, config, props, cache) {
    let prop = isArray(props) ? props[0] : props,
        isExpression = isExpressionType(config[prop]),
        __viewflow =  cache && isExpression ? cache.__tokens__ || cache.cache.__tokens__ : null,
        obj = isExpression ? config[prop].value || defaultExpressionValueAccessor : config[prop],
        __args = isExpression ? getExpressionArgs(config[prop], __viewflow, args) : args,
        val = getVal(isArray(props) ? props.reduce((prev, cur) => prev || config[cur], null) : obj, __args, node);
    // console.log(node);
    node.zc_cache = node.zc_cache || {};
    if (defined(val) || val === NULL) {//vall === null is checked since in FEATURESEVENTS_LEGEND_6_1_2, even when border disbaled for legend, border is got. 
        node.zc_cache[prop] = val;
    }
    // console.log(prop, val, node.zc_cache[prop], getVal(_defaultMarkValue[prop], args, node));
    return pick(val, node.zc_cache[prop], getVal(_defaultMarkValue[prop], args, node));
}

function isExpressionType(value){
    return isObject(value) && value._type == 'expression';
}
function getExpressionValue(node, args, config, prop, viewflow){
    let __args = getExpressionArgs(config, viewflow, args),
     val = getVal(config.value || defaultExpressionValueAccessor, __args, node);

    return val;
}

function defaultExpressionValueAccessor(d, i, args){
    let _scales = defined(args.scales),
        _columns = defined(args.columns),
        _reacts = defined(args.reacts),
        _data = d.data || d; //for pie layout it will return d.data
    if(_scales && _columns){
        return args.scales.component(_data[args.columns['dataindex']]);
    }else if(_columns && !_scales){
        return _data[args.columns['dataindex']];
    }else if(_reacts){
        return args.reacts.value;
    }
}
function getExpressionArgs(config, viewflow, _default){
    let allowedParserObjects = ['scales', 'columns', 'bound', 'transform', 'reacts'],
        parammsObjMap = {
            'transform': 'processedData'
        },
        values = {},
        userdefinedArgs = config.args;
    if(userdefinedArgs){
        allowedParserObjects.map(d => {
            let parserId = userdefinedArgs[d];
            if(defined(parserId)){
                let d1 = parammsObjMap[d] || d;
                values[d] = isArray(parserId) ? parserId.map(x => viewflow[d1][x]) : viewflow[d1][parserId];
            }
        })
    }
    
    return [..._default ,values];
}

function mapForExpressionParams(cache){
    let map = d3_map();
    let default_store = cache.__tokens__ || cache.cache.__tokens__,
        default_key = 

    map.set('scales', [default_store, 'scales']);
    map.set('columns', [default_store, 'columns']);
    map.set('bound', [default_store, 'bound']);
    map.set('scales', [default_store, 'scales']);
    map.set('transform', [cache, 'processedData']);

}

export { cssToJsObject, getMarkValue, getExpressionValue, isExpressionType, defaultExpressionValueAccessor };
