// @zohocharts/zc/src/legend/size-legend/nested/size-nested.js
//
import { Registry } from '../../../core/zc-registry';
import { LegendRegistry } from '../../legend-registry';
import { SizeLegend } from '../size-legend';
import { SizeNestedParser } from './size-nested-parser';
import { isOuterTickOrient } from './size-nested-axis-parser';
import { $ZCG } from '../../../core/zcglobal';
import { mathMax } from '../../../core/zc-constants';
import { bbox } from '../../../utils/zcutils';
import { getMaxToResize } from '../../legend-bound';
import { getTextHeight } from '../../../label/font-helpers';
import { renderDom } from '../../../mark/dom-render';

class SizeLegendNested extends SizeLegend {
    create() {
        this.encode = SizeNestedParser(this.spec, this.viewScaleComponent());
    }

    resize(width, height) {
        if (!width && !height) {
            return;
        }

        let { w, h } = getMaxToResize(width, height, this.spec.bound),
            children = this.encode.children,
            //bbox
            backgroundBox = bbox(children[0].getElement()),
            nestedBox = bbox(children[1].children[1].getElement());

        w -= width ? mathMax(backgroundBox.width - nestedBox.width, 0) : 0;
        h -= height ? mathMax(backgroundBox.height - nestedBox.height, 0) : 0;

        this.updateScale(w, h);
        super.resize(width, height);
    }

    draw(element) {
        this.element = element;
        renderDom(this.encode, element);

        let { config, bound } = this.spec,
            { maxWidth, maxHeight } = config,
            w = bound.maxWidth,
            h = bound.maxHeight,
            bb = bbox(element),
            bb_width = bb.width,
            bb_height = bb.height;

        //ZC2184, ZC2182
        // If user defined maxWidth or maxHeight is present adjust
        if ((maxWidth || maxHeight) && (bb_width - w > 1 || bb_height - h > 1)) {
            w -= maxWidth && bb_width > w ? bb_width - w : 0;
            h -= maxHeight && bb_height > h ? bb_height - h : 0;

            this.updateScale(w, h);
            this.create();
            renderDom(this.encode, element);
        }
    }
}

SizeLegendNested.adjustFactor = (config, w, h, globalFont) => {
    let adjust = [0, 0],
        { maxWidth, maxHeight } = config,
        axisConfig = config.nested.axis;

    if (axisConfig.show) {
        if (!maxHeight) {
            adjust[1] += getTextHeight([axisConfig.ticklabel, globalFont, $ZCG]);
        }

        if (!maxWidth) {
            adjust[0] += w * (isOuterTickOrient(axisConfig) ? 0.3 : 0.2);
        }
    }

    return adjust;
};

SizeLegendNested.encode = 'z';
SizeLegendNested.view = 'nested';
SizeLegendNested.domMode = 'svg';

export { SizeLegendNested };
Registry.setComponent('sizeLegendNested', SizeLegendNested);
LegendRegistry.add(SizeLegendNested);
