// @zohocharts/zc/src/legend/size-legend/nested/size-nested-parser.js
//
import { LegendTitleSvgParser } from '../../title/legend-title-svg-parser';
import { _LegendBackgroundSvgParser } from '../../background/legend-background-svg-parser';
import { LegendContentWrapParser } from '../legend-content-wrap-parser';
import { NestedAxisParser } from './size-nested-axis-parser';
import { NestedParser } from '../../../view/nested/nested-parser';
import { Mark } from '../../../mark/mark-utils';
import { mathMax } from '../../../core/zc-constants';
import { d3_values, bbox, pick, parseShortHandValue, setPropVal } from '../../../utils/zcutils';

function SizeNestedParser(spec, viewScaleComponent) {
    let { titleSpec, viewSpec, axisSpec, backgroundSpec, axisEnd } = helper(spec, viewScaleComponent); // TODO: implement axisEnd

    let groupMark = Mark('group', '.legendgroup'),
        backgroundMark = _LegendBackgroundSvgParser(backgroundSpec),
        contentMark = LegendContentWrapParser();

    // Title
    let titleMark = LegendTitleSvgParser(titleSpec);
    setPropVal(titleMark, 'as.height', 'title_height');

    // Nested view
    let nestedMark = NestedParser(viewSpec);
    setPropVal(nestedMark, 'config.y', { _type: 'as', as: 'title_height' });

    // Axis
    let axisMark = NestedAxisParser(axisSpec);
    axisMark.onEnd = () => axisEnd(axisMark, nestedMark, titleMark); // TODO: Replace by 'as'

    contentMark.children = [titleMark, nestedMark, axisMark];
    groupMark.children = [backgroundMark, contentMark];

    return groupMark;
}

function helper(spec, viewScaleComponent) {
    let { config, bound, format } = spec,
        { scale: viewScale, bound: viewBound, ticks, labels } = viewScaleComponent,
        //
        nestedConfig = config.nested,
        markerConfig = nestedConfig.marker,
        axisConfig = nestedConfig.axis,
        padding = pick(nestedConfig.padding, 3),
        parsedPadding = parseShortHandValue(padding);

    let titleSpec = {
            data: [1],
            config: config.title,
            bound
        },
        viewSpec = {
            bound: viewBound,
            data: d3_values(ticks).reverse(),
            scale: viewScale,
            config: markerConfig
        },
        axisSpec = {
            bound: viewBound,
            data: labels || ticks,
            scale: viewScale,
            config: axisConfig,
            align: { hAlign: markerConfig.hAlign, vAlign: markerConfig.vAlign },
            format,
            shape: markerConfig.shape,
            availableBound: {
                maxWidth: bound.maxWidth - parsedPadding.left - parsedPadding.right,
                maxHeight: bound.maxHeight - parsedPadding.top - parsedPadding.bottom
            }
        },
        backgroundSpec = {
            data: [1],
            config: config,
            margin: padding
        },
        axisEnd = (axisEncode, nestedEncode, titleEncode) => {
            let axisEle = axisEncode.getElement(),
                titleEle = titleEncode.getElement(),
                nestedEle = nestedEncode.getElement(),
                axisBox = bbox(axisEle),
                titleBox = bbox(titleEle),
                y = mathMax(-axisBox.y, 0) + titleBox.height + padding;

            axisEle.attr('transform', `translate(0,${y})`);
            nestedEle.attr('transform', `translate(0,${y})`);
        };

    return { titleSpec, viewSpec, axisSpec, backgroundSpec, axisEnd };
}

export { SizeNestedParser };
