// @zohocharts/zc/src/legend/size-legend/nested/size-nested-axis-parser.js
//
import {
    TOP_ALIGN,
    BOTTOM_ALIGN,
    LEFT_ALIGN,
    RIGHT_ALIGN,
    CENTER_ALIGN,
    END_ALIGN,
    MIDDLE_ALIGN,
    OUTER_ALIGN
} from '../../../core/align-constants';
import { NESTED_AXIS_CLASSNAME, NESTED_AXIS_TICKLABEL_CLASSNAME, NESTED_AXIS_TICKMARK_CLASSNAME } from '../../legend-constants';
import { Mark } from '../../../mark/mark-utils';
import { translateString } from '../../../utils/svg-transforms';
import { pick, parseShortHandValue, toArray, objectValues, setPropVal } from '../../../utils/zcutils';
import { handleTicklabels } from './size-nested-ticklabel-handler';

let CENTER_CENTER_ALIGN = 'center-center';

// TODO: check
function NestedAxisParser(spec) {
    let { data, config, bound, scale, align, format, shape = 'square', availableBound } = spec,
        //Config
        { ticklabel: lableConfig, tickmark: lineConfig, show } = config,
        outerOrient = isOuterTickOrient(config),
        //
        tickMarkParser = show && outerOrient && parseTickMark(lineConfig, bound, scale, align, shape),
        tickLabelParser = show && parseTickLabel(lableConfig, bound, scale, align, format, lineConfig, shape, outerOrient),
        tickLabelEnd = (labelEle, lineEle) => handleTicklabels(labelEle, lineEle, availableBound, bound, tickLabelParser.config);

    let groupMark = Mark('group', `.${NESTED_AXIS_CLASSNAME}`);

    if (!show) {
        return groupMark;
    }

    // Tick mark
    let lineGroupMark = Mark('group', `.${NESTED_AXIS_TICKMARK_CLASSNAME}`);
    if (tickMarkParser) {
        let lineMark = { mark: 'line', data: data, config: tickMarkParser.config };
        setPropVal(lineMark, 'properties.styles', tickMarkParser.styles);
        lineGroupMark.children = [lineMark];
    }

    // Tick label
    let labelGroupMark = Mark('group', `.${NESTED_AXIS_TICKLABEL_CLASSNAME}`);
    if (labelGroupMark) {
        let labelMark = { mark: 'text', data: data, config: tickLabelParser.config };
        setPropVal(labelMark, 'properties.styles', tickLabelParser.styles);
        labelGroupMark.children = [labelMark];
        labelGroupMark.onEnd = ele => tickLabelEnd(ele.selectAll('text'), lineGroupMark.getElement().selectAll('line'));
    }

    groupMark.children = [lineGroupMark, labelGroupMark];

    return groupMark;
}

function parseTickMark(config, bound, scale, align, shape) {
    let { hAlign, vAlign } = getAxisAlignment(align),
        { color, size = 3 } = config,
        position = linePoints(bound, scale, hAlign, vAlign, size, shape);

    return {
        config: { strokeColor: color, ...position, ...config },
        styles: { display: null, fillColor: 'none' }
    };
}

function parseTickLabel(config, bound, scale, align, format, tickMarkConfig, shape, outerOrient) {
    let { hAlign, vAlign } = getAxisAlignment(align),
        x,
        y,
        dy,
        anchor = null,
        margin = pick(config.margin, 2),
        isHorizontal = vAlign === CENTER_ALIGN;

    if (outerOrient) {
        let tickMarkSize = pick(tickMarkConfig.size, 3),
            newMargin = mergeShortHandValues(tickMarkSize, margin);

        ({ x2: x, y2: y } = linePoints(bound, scale, hAlign, vAlign, newMargin, shape));
        dy = isHorizontal ? '1em' : '0.3em';
        anchor = isHorizontal ? MIDDLE_ALIGN : null;
    } else {
        [x, y] = startXY(bound, scale, hAlign, vAlign, margin, shape);
        dy = _dy(vAlign);
        anchor = _anchor(hAlign);
    }

    return {
        config: {
            text: format,
            transform: d => translateString([x(d), y(d)]),
            anchor,
            dy,
            ...config
        },
        styles: { display: null }
    };
}

function _x(bound, scale, hAlign, margin = 0) {
    let parsedMargin = parseShortHandValue(margin),
        w = bound.width;

    switch (hAlign) {
        case LEFT_ALIGN:
            return d => scale(d) + parsedMargin.left;
        case RIGHT_ALIGN:
            return d => w - scale(d) - parsedMargin.right;
        default:
            return () => w / 2;
    }
}

function _y(bound, scale, vAlign, margin = 0) {
    let parsedMargin = parseShortHandValue(margin),
        h = bound.height;

    switch (vAlign) {
        case TOP_ALIGN:
            return d => scale(d) + parsedMargin.top;
        case BOTTOM_ALIGN:
            return d => h - scale(d) - parsedMargin.bottom;
        case CENTER_CENTER_ALIGN:
            return d => (h - scale(d)) / 2 - parsedMargin.bottom;
        default:
            return () => h / 2;
    }
}

function _anchor(hAlign) {
    switch (hAlign) {
        case LEFT_ALIGN:
            return null;
        case RIGHT_ALIGN:
            return END_ALIGN;
        default:
            return MIDDLE_ALIGN;
    }
}

function _dy(vAlign) {
    switch (vAlign) {
        case TOP_ALIGN:
            return '1em';
        case BOTTOM_ALIGN:
        case CENTER_CENTER_ALIGN:
            return '-0.3em';
        default:
            return '0.3em';
    }
}

function startXY(bound, scale, hAlign, vAlign, margin, shape, outerOrient) {
    let x = _x(bound, scale, hAlign, margin),
        y = _y(bound, scale, vAlign, margin),
        startX = x,
        startY = y;

    if (shape === 'square') {
        if (outerOrient) {
            if (vAlign === CENTER_ALIGN) {
                startY = d => y(d) + scale(d) / 2;
            } else if (hAlign === CENTER_ALIGN) {
                startX = d => x(d) + scale(d) / 2;
            } else if (hAlign === RIGHT_ALIGN) {
                startX = d => x(d) + scale(d);
            }
        }
    } else if (vAlign !== CENTER_ALIGN) {
        if (hAlign === LEFT_ALIGN) {
            startX = d => x(d) - scale(d) / 2;
        } else if (hAlign === RIGHT_ALIGN) {
            startX = d => x(d) + scale(d) / 2;
        }
    }

    return [startX, startY];
}

function linePoints(bound, scale, hAlign, vAlign, size, shape) {
    size = parseShortHandValue(size);

    let _startXY = startXY(bound, scale, hAlign, vAlign, 0, shape, true),
        x1 = _startXY[0],
        y1 = _startXY[1],
        x2 = () => bound.width + size.left,
        y2 = () => bound.height + size.top;

    if (vAlign === CENTER_ALIGN) {
        x2 = x1;
    } else {
        y2 = y1;
    }
    return { x1, y1, x2, y2 };
}

function getAxisAlignment(align) {
    let { hAlign, vAlign } = align;
    if (vAlign === CENTER_ALIGN && hAlign === CENTER_ALIGN) {
        vAlign = CENTER_CENTER_ALIGN;
    }

    return { vAlign, hAlign };
}

function mergeShortHandValues() {
    let values = toArray(arguments).slice(0),
        sides = { top: 0, right: 0, bottom: 0, left: 0 };

    values.map(_ => {
        let parsed = parseShortHandValue(_);
        for (let side in sides) {
            sides[side] += parsed[side];
        }
    });

    return objectValues(sides).join(' ');
}

function isOuterTickOrient(config) {
    let tickOrient = config.tickOrient;
    return tickOrient === OUTER_ALIGN;
}

export { NestedAxisParser, isOuterTickOrient };
