// @zohocharts/zc/src/legend/size-legend/nested/size-nested-ticklabel-handler.js
//
import { BOTTOM_ALIGN } from '../../../core/align-constants';
import { trimText } from '../../../utils/zcsvgutils';
import { bbox } from '../../../utils/zcutils';
import { d3_select } from '../../../core/d3-constants';
import { doOverlap } from '../../../utils/geometry-utils';
import { getSvgTransform } from '../../../utils/svg-transforms';

function handleTicklabels(ele, tickMarkEle, availableBound, bound, labelParser) {
    if (ele.empty()) {
        return;
    }

    let labelBounds,
        e = ele.nodes(),
        lineElements = tickMarkEle.nodes(),
        edge = _edge(availableBound, bound, labelParser.anchor),
        translateFunction = labelParser.translateFunction,
        { _margin } = labelParser,
        handler = (ele, i) => {
            d3_select(ele).style('display', 'none');
            d3_select(lineElements[i]).style('display', 'none');
            labelBounds[i] = { x1: 0, y1: 0, x2: 0, y2: 0 };
        };

    // TODO remove: minmaxPriority - DOCEVENTS_LEGEND_18 no priority for color band axis
    if (labelParser.minmaxPriority !== false) {
        e.splice(1, 0, e.splice(e.length - 1, 1)[0]);
        lineElements.splice(1, 0, lineElements.splice(lineElements.length - 1, 1)[0]);
    }

    labelBounds = getLabelBounds(e, translateFunction);

    handleOverflow(e, edge, handler, translateFunction, _margin);
    handleOverlap(e, labelBounds, handler);
}

// TODO: reuse
function handleOverlap(ele, eleBounds, handler) {
    ele.forEach((e, i) => {
        for (var index = i - 1, overlap = false; index >= 0 && !overlap; index--) {
            overlap = doOverlap(eleBounds[index], eleBounds[i]);
            if (overlap) {
                handler(e, i);
            }
        }
    });
}

// TODO: reuse
function handleOverflow(ele, edge, handler, translateFunction, margin = { left: 0, right: 0, bottom: 0, top: 0 }) {
    let pattern = '..';

    ele.forEach((node, i) => {
        let bb = bbox(node),
            translate = getTranslate(node, translateFunction),
            textBound = {
                left: bb.x + translate[0],
                top: bb.y + translate[1],
                width: bb.width,
                height: bb.height
            };

        //TODO: ticklabel.maxWidth
        trimText(node, node.textContent, textBound, edge, BOTTOM_ALIGN, pattern, margin);

        if (node.textContent === pattern) {
            d3_select(node).text('');
        }

        if (!node.textContent) {
            handler(node, i);
        }
    });
}

function getLabelBounds(elements, translateFunction) {
    return elements.map(_ => getLabelBound(_, translateFunction));
}

function getLabelBound(ele, translateFunction) {
    var bb = bbox(ele),
        translate = getTranslate(ele, translateFunction);

    var bound = {
        x1: bb.x + translate[0],
        y1: bb.y + translate[1],
        width: bb.width,
        height: bb.height
    };

    bound.x2 = bound.x1 + bound.width;
    bound.y2 = bound.y1 + bound.height;

    return bound;
}

function _edge(availableBound, bound, textAnchor, margin = { left: 0, right: 0, bottom: 0, top: 0 }) {
    let { x = 0, y = 0, maxWidth, maxHeight } = availableBound,
        //
        top = y + margin.top,
        bottom = top + maxHeight - margin.bottom,
        left = x + margin.left,
        right = left + maxWidth - margin.right;

    switch (textAnchor) {
        case 'end': {
            let r = maxWidth - bound.width;
            left -= r;
            right -= r;
            break;
        }
        case 'middle': {
            let r = (maxWidth - bound.width) / 2;
            left -= r;
            right -= r;
            break;
        }
    }

    return { top, right, bottom, left };
}

function getTranslate(ele, translateFunction) {
    return translateFunction ? translateFunction(ele) : getSvgTransform(ele).translate;
}

export { handleTicklabels, handleOverlap, getLabelBounds };
