// @zohocharts/zc/src/scale-parser/axis/correction/bubble-correction-parser.js
//$Id$
import { d3_extent, d3_scaleOrdinal } from '../../../core/d3-constants';
import { scale } from '../../../core/main-zc-init';
import { getContinuousScale } from '../../../axis/scale';
import { getKey } from './correction-parser-helpers';
import { DataProcessor } from '../../../core/data-processor';
import { getChartTypeByIndex, isBubbleChartType } from '../../../core/charttype-helpers';
import { _modularized } from '../../../core/zcconf';
import { mathMin } from '../../../core/zc-constants';

// TODO: array of series data in font precalculation. data to be a function. user cannot iterate and give the series data
export function bubbleCorrectionParser(encodeType, index, chart) {
    let key = getKey('bubble', encodeType, index),
        operator = [],
        dataSet = chart.dataset,
        processor = chart.dataObject,
        encodeData = processor.getEncode(encodeType, index),
        seriesData = encodeData.seriesData,
        helpers = DataProcessor.helpers,
        margin = 0;

    if (!processor.isBubbleEnabled) {
        return operator;
    }

    seriesData.forEach((series, seriesIndex) => {
        var disabled = series.disabled,
            yaxiscolumnorder = helpers.getParsedYAxisOrder(series.yaxiscolumnorder),
            isCatNullX = helpers.isCatNull(dataSet, 'x'),
            isCatNullY = helpers.isCatNull(dataSet, 'y', yaxiscolumnorder),
            condition = d => !(!d || d.isDummy || isCatNullX(dataSet.getX(d)) || isCatNullY(dataSet.getY(d, yaxiscolumnorder)));

        if (
            _modularized.scaleCorrection ? !disabled && isBubbleChartType(getChartTypeByIndex(chart, null, seriesIndex).value) : !disabled
        ) {
            series.data.forEach(subgroup => {
                operator.push({
                    _type: 'marker',
                    data: subgroup.filter(condition),
                    size: [d => processor.getXYvalue(null, null, null, null, null, d, 'z'), margin],
                    position: d => (encodeType === 'x' ? dataSet.getX(d) : dataSet.getY(d, yaxiscolumnorder)),
                    scale: getScaleForBubbleCorrection,
                    key
                });
            });
        }
    });

    return operator;
}

function getScaleForBubbleCorrection(_scale) {
    let _chart = _scale.chartObj,
        _config = _scale.config,
        domain,
        range,
        scaleObj;

    if (_scale.config.domain.hasDefaultMin) {
        _scale.config.tempDomain = [0, _scale.config.tempDomain[1]];
        _scale.config.extremes = [0, _scale.config.extremes[1]];
    }
    if (_scale.config.type == 'ordinal') {
        domain = _config.tempDomain || _config.domain;
        range = d3_extent(_config.tempRange); // if (range[0] > range[1]) { range = [range[1], range[0]]; }

        var d3_outerPadding = scale.getOuterPadding(range, _config.domain.length, _config.padding, _config.outerPadding);
        d3_outerPadding = mathMin(d3_outerPadding, 1);
        scaleObj = d3_scaleOrdinal()
            .domain(domain)
            .range(range)
            .paddingInner(_config.padding)
            .paddingOuter(d3_outerPadding);
    } else {
        domain = _config.tempDomain;
        range = [0, _chart.plotarea[_config.range.operators.plotBound.dimension]];
        scaleObj = getContinuousScale(_chart.dataset, _scale.encode, _scale.key)()
            .domain(domain)
            .range(range);
    }
    return scaleObj;
}
