// @zohocharts/zc/src/behaviours/legend-brush/size-slider/legend-size-slider-brush-handler-mappings.js
import { LegendSizeSliderBrushListener } from '../../listeners/legend/legend-size-slider-brush-listener';
import { LegendBrushDataFinder } from '../../data-selection/legend/legend-brush-data-finder';
import { SLIDER_LEGEND_TYPE } from '../../../legend/legend-type-constants';
import { Registry } from '../../../core/zc-registry';
import { d3_set } from '../../../core/d3-constants';
import { validateNullVal } from '../../../utils/zcutils';
import { getDisabledOptions } from '../../../legend/events/legend-event-helpers';
import { getHandles } from './handles';
import { brushTooltip } from './brush-tooltip';
import { BrushClip } from './brush-clip';
import { handleShadow } from '../../visual-effects/brush-handles/handle-shadow';
// import { updateTrackerClip } from '../../../legend/colorband/colorband-track';
import { isSupported } from '../../../legend/events/legend-event-helpers';

import { styleHandles } from '../../visual-effects/legend/size-slider/style-handles';
import { ApplyBrushSelectionOnPlot } from '../../visual-effects/legend/size-slider/update-plot-by-selection';
import { showTooltip } from '../../visual-effects/legend/size-slider/show-tooltip';
import { hideTooltip } from '../../visual-effects/legend/size-slider/hide-tooltip';
import { HideHighlight } from '../../visual-effects/legend/size-slider/hide-highlight';
import { PreventDefault } from '../../visual-effects/legend/size-slider/slider-prevent-default';
import { SnapRange } from '../../visual-effects/legend/size-slider/snap-range';
import { styleConnector } from '../../visual-effects/legend/size-slider/style-connector';
import { ForceHideTooltip } from '../../visual-effects/legend/size-slider/force-hide-tooltip';
import { ClearCache } from '../../../legend/events/visual-effects/clear-cache';
import { AddClipToHandles } from '../../visual-effects/legend/size-slider/add-clip-to-handles';
import { StyleTrack } from '../../visual-effects/legend/size-slider/style-track';

let _legendSizeSliderBrushHandlers = {
    LISTENER: {
        start: LegendSizeSliderBrushListener,
        brush: LegendSizeSliderBrushListener,
        end: LegendSizeSliderBrushListener
    },
    DATA_FINDER: {
        start: LegendBrushDataFinder,
        brush: LegendBrushDataFinder,
        end: LegendBrushDataFinder
    },
    BEHAVIOUR: {
        start: [PreventDefault],
        brush: [ForceHideTooltip, styleHandles, ApplyBrushSelectionOnPlot, showTooltip, HideHighlight, styleConnector, StyleTrack],
        end: [ClearCache, hideTooltip, SnapRange, AddClipToHandles]
    },
    LISTENER_TYPES: d3_set(['start', 'brush', 'end']),
    SUPPORTED: isSupported(SLIDER_LEGEND_TYPE),
    // UPDATE: function() {},
    INIT: function(behaviour) {
        let behaviouralEvents = this,
            chart = behaviouralEvents.chart,
            behaviourInstance = behaviour.instance,
            legendComponent = behaviourInstance.options.component,
            { initialExtent, config, handles: handleOptions } = legendComponent.getBrushOptions(),
            handles = getHandles(behaviour),
            eventEnabled = isEventEnabled(config);

        behaviour.initialized = false;
        behaviourInstance.setSelection(initialExtent);
        let handleElement = behaviourInstance.addCustomHandle(handles);
        behaviourInstance.tooltip = brushTooltip(behaviour);

        // Shadow
        let filterConfig = getDisabledOptions(config);
        handleShadow(filterConfig, handles, handleElement, chart);

        // Clip
        BrushClip(behaviour, chart);
        AddClipToHandles.call(this, behaviour);
        // updateTrackerClip(brushOptions, behaviourInstance.parent_group, behaviourInstance, chart);

        // Visibility for custom handles
        // behaviourInstance.customHandles.style('visibility', eventEnabled || handleOptions ? null : 'hidden');

        if (!eventEnabled) {
            behaviourInstance.customHandles.attr('pointer-events', 'none');
            behaviourInstance.destroy();
        }

        behaviour.initialized = true;
    }
};

function isEventEnabled(config) {
    let { events: eventsConfig, filter } = config,
        eventDisabled = eventsConfig && validateNullVal(eventsConfig.enabled, true) == false;

    return !eventDisabled && filter.enabled;
}

Registry.setComponent('_legendSizeSliderBrushHandlers', _legendSizeSliderBrushHandlers);

export { _legendSizeSliderBrushHandlers };
