// @zohocharts/zc/src/behaviours/legend-brush/size-slider/legend-size-slider-brush-arguments.js
import { SLIDER_VIEW_CLASSNAME } from '../../../legend/legend-constants';
import { Registry } from '../../../core/zc-registry';
import { TRANSPARENT } from '../../../core/zc-constants';

/**
 *
 * @param chart
 * @return {BrushD3Options}
 */
export function _getLegendSizeSliderBrushArguments(chart) {
    let legendInstance = chart.legends.get('z-0'),
        { component, element } = legendInstance,
        { brushOptions } = component,
        { type, dimension } = brushOptions,
        //
        extent = getExtent(dimension, brushOptions),
        BrushD3Options = Registry.getComponent('BrushD3Options');

    return new BrushD3Options({
        type,
        extent,
        component,
        behaviour: 'resizeToMouse',
        element: getTrackerElement(element),
        styles: {
            selection: {
                fillColor: TRANSPARENT,
                strokeColor: TRANSPARENT,
                fillOpacity: 1
            },
            track: {
                fillColor: TRANSPARENT
            }
        }
    });
}

/**
 * Returns tracker element
 * @param  {D3Selectiion} container
 * @return {D3Selectiion}
 */
function getTrackerElement(container) {
    return container.selectAll(`g.${SLIDER_VIEW_CLASSNAME}`);
}

function getExtent(dimension, options) {
    let [width, height] = dimension,
        { brushScale, isHorizontal, ranges } = options,
        bandWidth = ranges && ranges.length === 1 ? 0 : brushScale.bandwidth ? brushScale.bandwidth() : 0,
        bandWidthBy2 = bandWidth / 2;

    if (isHorizontal) {
        return [[bandWidthBy2, 0], [width - bandWidthBy2, height]];
    } else {
        return [[0, bandWidthBy2], [width, height - bandWidthBy2]];
    }
}

Registry.setComponent('_getLegendSizeSliderBrushArguments', _getLegendSizeSliderBrushArguments);
