// @zohocharts/zc/src/legend/color-legend/band/band-parser.js
//
import { Mark } from '../../../mark/mark-utils';
import { d3_scalePoint, d3_scaleLinear } from '../../../core/d3-constants';
import { setPropVal } from '../../../utils/zcutils';

function BandParser(spec) {
    let { config, dimension, patternSVG, defsPath } = spec;

    let bandMark = Mark('_rect', '.zc-band');
    setPropVal(bandMark, 'properties.styles.pointer-events', 'visible');
    bandMark.config = {
        ...config,
        width: dimension[0],
        height: dimension[1],
        fillColor: {
            _type: 'gradient',
            gradient: 'linear',
            context: patternSVG,
            dimension: [0, dimension[0]],
            ...helper(spec),
            defsPath
        }
    };

    return bandMark;
}

function helper(spec) {
    let { reversed, isHorizontal, scale, config } = spec,
        userRanges_length = config.ranges && config.ranges.length,
        scale_domain,
        scale_range,
        offsetScale,
        bound;

    if (reversed) {
        bound = { x1: isHorizontal ? 100 : 0, x2: null, y1: null, y2: isHorizontal ? 0 : 100 };
    } else {
        bound = { x1: null, x2: isHorizontal ? 100 : 0, y1: isHorizontal ? 0 : 100, y2: null };
    }

    if (userRanges_length) {
        scale_range = scale.range().filter(d => d !== null);
        let temp_domain = [],
            temp_range = [];
        for (let i = 0; i < userRanges_length; i++) {
            temp_domain.push(i, i + 1);
            temp_range.push(scale_range[i], scale_range[i]);
        }
        scale_domain = temp_domain;
        scale_range = temp_range;

        offsetScale = d3_scalePoint()
            .domain(scale_domain)
            .range([0, 100]);
    } else {
        scale_domain = scale.domain();
        scale_range = scale.range();

        if (scale_domain[0] === scale_domain[1]) {
            //Test_696.json
            scale_range = [scale_range[0], scale_range[0]];
        }

        offsetScale = d3_scaleLinear()
            .domain([scale_domain[0], scale_domain[scale_domain.length - 1]])
            .range([0, 100]);
    }

    return { ...bound, stopColor: scale_range, stopOffset: scale_domain.map(d => `${offsetScale(d)}%`) };
}

export { BandParser };
