// @zohocharts/zc/src/legend/color-legend/band/band-axis-parser.js
//
import { BOTTOM_ALIGN, TOP_ALIGN, RIGHT_ALIGN, LEFT_ALIGN, START_ALIGN, END_ALIGN } from '../../../core/align-constants';
import { ORDINAL_DATATYPE } from '../../../core/datatype-constants';
import { Mark } from '../../../mark/mark-utils';
import { d3axis } from '../../../axis/d3-axis';
import { mathAbs } from '../../../core/zc-constants';
import { d3_select, d3_scaleLinear, d3_scaleOrdinal, d3_scalePoint, d3_merge } from '../../../core/d3-constants';
import { pick, defined } from '../../../utils/zcutils';
import { addSVGtitleEle } from '../../../utils/zcsvgutils';
import { getLegendRanges } from '../../../axis/colorscale-helpers';
// import { legend_helpers } from '../../legend-helper';
import { legend_helpers } from '../../../core/main-zc-init';
import { bindEventHandlerForTicks } from '../../../axis/axis-events';
import { applyAxislineProperties } from '../../legend-coloraxis';

function BandAxisParser(spec) {
    let { axisSpec, updateAxisLine, trimText } = helper(spec),
        { config, meta, callbackArgs = [] } = spec;

    let groupMark = Mark('group', '.zc-band-axis');

    if (config.show === false) {
        return groupMark;
    }

    let drawAxis = ele => {
        ele.call(d3axis, axisSpec, config);
        updateAxisLine(ele.selectAll('path.domain'));
        trimText(ele);
        bindEventHandlerForTicks(callbackArgs[0], null, getLabelEle(ele), null, config.ticklabel, { data: [meta] }, 'clr');

        let labelEle = getLabelEle(ele);
        labelEle.style('pointer-events', 'visible');
    };
    let axisGroupMark = Mark('group');
    axisGroupMark.onEnd = drawAxis;

    groupMark.children = [axisGroupMark];
    return groupMark;
}

function helper(spec) {
    let { scale, config, format = d => d, orient, dimension, dataType, reversed, isHorizontal, bandConfig = {} } = spec, // TODO: format
        { viewBound, availableBound, globalFont = {} } = spec,
        {
            tickmark: tickMarkconfig,
            ticklabel: tickLabelConfig,
            axisline: axisLineConfig,
            grid: gridConfig,
            minmaxlabels: minmaxLabelConfig
        } = config,
        axisLineSize = axisLineConfig.show ? pick(axisLineConfig.strokeWidth, 1) : 0,
        { labels } = bandConfig,
        { tooltip } = tickLabelConfig,
        tickFormat = function(d, i) {
            if (labels && labels.length) {
                var labelIndex = legend_helpers.getLabelIndex(d, bandConfig, dataType);
                if (defined(labelIndex)) {
                    let _text = labels[labelIndex];
                    this.__formattedValue = _text;
                    return d3_select(this).text(_text);
                }
            }

            let _text = format(d);
            this.__formattedValue = _text;
            d3_select(this).text(_text);
        },
        //
        dummyChartObj = {
            dataObject: {},
            systemConf: { canvas: { ...globalFont }, chart: { plot: { border: {} } } }
        },
        dummyAxisObj = { ticks: { isHierarchialLevelType: () => {} } },
        addTooltip = function() {
            addSVGtitleEle(d3_select(this), this.__formattedValue);
        },
        scale_domain = scale.domain(),
        range = isHorizontal ? [0, dimension[0]] : [dimension[1], 0],
        axisData = minmaxLabelConfig.show && minmaxLabelConfig.excludeIntervals ? [] : scale_domain,
        posScale;

    if (reversed) {
        range.reverse();
    }

    if (bandConfig.ranges && bandConfig.ranges.length) {
        scale_domain = getLegendRanges(bandConfig, dataType);
        if (bandConfig.type === 'continuous' && dataType !== ORDINAL_DATATYPE) {
            scale_domain = d3_merge(scale_domain);
            if (bandConfig.minRange !== scale_domain[0]) {
                scale_domain.splice(0, 0, bandConfig.minRange);
            }

            posScale = d3_scalePoint()
                .domain(scale_domain)
                .range(range);
        } else {
            posScale = d3_scaleOrdinal()
                .domain(scale_domain)
                .range(range)
                .padding(0);
        }
    } else {
        if (axisData[0] === axisData[1]) {
            //Test_694.json
            var rangeDiff = mathAbs(range[0] - range[1]);
            range = [rangeDiff / 2, rangeDiff / 2];
        }

        posScale = d3_scaleLinear()
            .domain([scale_domain[0], scale_domain[scale_domain.length - 1]])
            .range(range);
    }
    posScale.origRange = range;
    axisData = minmaxLabelConfig.show && minmaxLabelConfig.excludeIntervals ? [] : scale_domain;

    let axisSpec = {
            orient,
            scale: posScale,
            tickFormat,
            tickValues: axisData,
            tickSizeInner: pick(tickMarkconfig.size, 6),
            tickSizeOuter: 0,
            gridSize: getMajorTicks(orient, dimension[0], dimension[1], axisLineSize),
            tickPadding: pick(tickLabelConfig.margin, 3),
            rotation: 0,
            renderTicklabel: true,
            renderGrid: true,
            isGridShow: gridConfig.show,
            chart: dummyChartObj, // d3axis requires chartObj
            axis: dummyAxisObj // d3axis requires axisObj
        },
        updateAxisLine = ele => applyAxislineProperties(ele, axisLineConfig, orient, isHorizontal, dimension),
        trimText = ele => {
            return;
            // let labelEle = getLabelEle(ele),
            //     lineEle = getLineEle(ele),
            //     _anchor = {
            //         anchor: getAnchor(orient),
            //         minmaxPriority: false, // TODO: remove
            //         _margin: { left: 5, right: 5, bottom: 5, top: 5 },
            //         translateFunction: ele => {
            //             let textTranslate = getSvgTransform(ele).translate,
            //                 gTranslate = getSvgTransform(ele.parentNode).translate;
            //
            //             return [textTranslate[0] + gTranslate[0], textTranslate[1] + gTranslate[1]];
            //         }
            //     };
            //
            // handleTicklabels(labelEle, lineEle, availableBound, viewBound, _anchor);
            //
            // //  ZC2424 after handle overflow, tooltip will not be shown
            // if (tooltip === true) {
            //     labelEle.each(addTooltip);
            // }
        };

    return { axisSpec, updateAxisLine, trimText };
}

function getMajorTicks(orient, w, h, axislineSize) {
    var x1, y1, x2, y2;
    x1 = y1 = x2 = y2 = 0;

    if (orient == BOTTOM_ALIGN) {
        y1 = -axislineSize;
        y2 = -h - axislineSize;
    } else if (orient == TOP_ALIGN) {
        y1 = axislineSize;
        y2 = h + axislineSize;
    } else if (orient == LEFT_ALIGN) {
        x1 = w + axislineSize;
        x2 = axislineSize;
    } else if (orient == RIGHT_ALIGN) {
        x1 = -w - axislineSize;
        x2 = -axislineSize;
    }

    return { x1, y1, x2, y2 };
}

function getAnchor(orient) {
    return orient === RIGHT_ALIGN ? START_ALIGN : orient === LEFT_ALIGN ? END_ALIGN : START_ALIGN;
}

function getLabelEle(ele) {
    return ele.selectAll('text.ticklabel');
}

function getLineEle(ele) {
    return ele.selectAll('line.minorTicks');
}

export { BandAxisParser };
