// @zohocharts/zc/src/legend/color-legend/band/brush-handle-parser.js
//
import { ORDINAL_DATATYPE } from '../../../core/datatype-constants';
import { BrushD3 } from '../../../behaviours/listeners/brush/brush-d3';
import { Mark } from '../../../mark/mark-utils';
import { setPropVal, getBorderRadius, splat, isArray } from '../../../utils/zcutils';
import { roundedRect } from '../../../utils/zcsvgutils';
import { handlerByType } from '../../events/visual-effects/size-slider/highlight-item';
import { getHighlightedItemFromMousePoint } from '../../events/data-selection/size-slider-data-finder';
import { getRange } from '../../colorband/colorband-brush-helpers';

function BrushHandleParser(spec) {
    let _helper = helper(spec),
        { isHorizontal } = spec,
        { show, handles, rectHandle } = _helper;

    let groupMark = Mark('group', '.zc-band-brush-handle');
    setPropVal(groupMark, 'properties.styles', { visibility: 'hidden', 'pointer-events': 'none' });

    if (!show) {
        return groupMark;
    }

    groupMark.onEnd = ele => {
        handles.forEach(_ => {
            _.shape = splat(_.shape); // 'triangleUp' shape is not array
            _.style = splat(_.style);
            let transform = _.style[0].transform
                .replace('translate(', '')
                .replace(')', '')
                .split(' ')
                .map(_ => +_ || 0) || [0, 0];

            if (isArray(_.shape[0])) {
                _.shape.forEach(s => s.push(rectHandle)); // 'vernier' shape FEATURESEVENTS_HEATMAP_9_1
                _.style.push({
                    transform: 'translate(' + (isHorizontal ? transform[0] : 0) + ' ' + (isHorizontal ? 0 : transform[1]) + ')'
                });
            } else {
                _.shape.push(rectHandle);
                _.style.push({
                    transform: 'translate(' + (isHorizontal ? transform[0] : 0) + ' ' + (isHorizontal ? 0 : transform[1]) + ')'
                });
            }
        }); // d3 brush adds rect
        BrushD3.ADD_HANDLE(handles, ele);
    };

    return groupMark;
}

function helper(spec) {
    let { initialExtent, config, isHorizontal, dimension, brushScale, ranges, singleDomainContinuous } = spec,
        { enabled, slider: sliderConfig } = config.filter,
        bandWidth = brushScale.bandwidth ? brushScale.bandwidth() : 0,
        handleSize = 6,
        selection = singleDomainContinuous
            ? initialExtent
            : ranges && ranges.length
            ? initialExtent.map(_ => getHighlightData(getHighlightedItemFromMousePoint([_, _], spec), spec))
            : getRange(spec, initialExtent);

    return {
        show: enabled,
        handles: selection.map(_ => handlerByType(sliderConfig, spec, _)),
        rectHandle: roundedRect(
            -handleSize / 2,
            -handleSize / 2,
            (isHorizontal ? 0 : dimension[0]) + handleSize,
            (isHorizontal ? dimension[1] : 0) + handleSize,
            getBorderRadius(0)
        )
    };
}

function getHighlightData(point, options) {
    let { dataType, ranges: userRanges } = options;

    if (dataType === ORDINAL_DATATYPE) {
        return userRanges[point][0];
    } else if (userRanges && userRanges.length) {
        return userRanges[point];
    } else {
        return point[0];
    }
}

export { BrushHandleParser };
