// @zohocharts/zc/src/legend/color-legend/band/color-band.js
//
import { LEFT_ALIGN, TOP_ALIGN, HORIZONTAL_ALIGN } from '../../../core/align-constants';
import { Registry } from '../../../core/zc-registry';
import { LegendRegistry } from '../../legend-registry';
import { ColorBandParser } from './color-band-parser';
import { ColorLegend } from '../color-legend';
import { mathMin } from '../../../core/zc-constants';
import { bbox, pick, each, getBoundingRect, mergeJSON } from '../../../utils/zcutils';
import { getLegendAlignment } from '../../legend-utils';
import { legend_helpers } from '../../../core/main-zc-init';
import { getBrushScale, getInitialExtent } from '../../legend-brush';
import { getColorAxisOptions } from '../../legend-continuous';
import { getLegendPosition } from '../../legend-utils';
import { getSvgTransform } from '../../../utils/svg-transforms';
import { getDefaultSliderOptions } from '../../../behaviours/visual-effects/brush-handles/brush-handle-options';
import { handleOverlapNOverlap } from '../../legend-coloraxis';

class ColorLegendBand extends ColorLegend {
    create() {
        this.brushOptions = this.getBrushOptions();
        this.encode = ColorBandParser({
            spec: this.spec,
            align: ele => {
                this.align(ele, this);
                // TODO: remove the following
                let spec = this.spec,
                    { config, chartObj, area } = spec,
                    axisConfig = getColorAxisOptions(config),
                    position = getLegendPosition(area, chartObj, null, ele, config),
                    axisGroup = ele.selectAll('g.zc-band-axis'),
                    _m = ele.selectAll('g.zc-legend-minmaxlabels').selectAll('text'),
                    axisXY = getSvgTransform(axisGroup).translate;
                handleOverlapNOverlap(axisGroup, axisXY, position, chartObj, spec, axisConfig, ele, _m);
                this.align(ele, this);
            },
            brushOptions: this.brushOptions
        });
        return this.encode;
    }

    align(ele, self) {
        var titleEle = getEle(self, 'LegendTitleSvg'),
            bandEle = getEle(self, 'Band'),
            axisEle = getEle(self, 'Axis'),
            minmaxLabelEle = getEle(self, 'MinmaxLabel'),
            brushEle = getEle(self, 'Brush'),
            brushHandleEle = getEle(self, 'BrushHandle'),
            // BBox
            titleBox = bbox(titleEle),
            bandBox = bbox(bandEle),
            minmaxLabelBox = bbox(minmaxLabelEle),
            axisBox = bbox(axisEle),
            brushBox = bbox(brushEle),
            //
            { isHorizontal, orient, dimension } = this.brushOptions,
            { config } = this.spec,
            bandConfig = config.band || config.colorBand,
            axisConfig = getColorAxisOptions(config),
            padding = 0, //pick(bandConfig.padding, 0),
            axisPadding = axisConfig.show ? pick(axisConfig.padding, 3) : 0,
            bandStrokeWidth = pick(bandConfig.strokeWidth, 1) / 2,
            { axisline: axisLineConfig } = axisConfig,
            axisLinsStrokeWidth = (axisLineConfig.show ? pick(axisLineConfig.strokeWidth, 1) : 0) - 0.5,
            x = 0,
            y = 0;

        translate([0, 0], titleEle);
        if (isHorizontal) {
            if (orient === TOP_ALIGN) {
                x = mathMin(0, axisBox.x);
                y = titleBox.height + axisBox.height + padding + bandStrokeWidth;

                translate([x, y + bandStrokeWidth + axisLinsStrokeWidth], bandEle, brushEle, brushHandleEle);
                translate([x, y], axisEle, minmaxLabelEle);
            } else {
                x = -mathMin(0, minmaxLabelBox.x);
                y = titleBox.height + padding;

                translate([x, y - brushBox.y], bandEle, brushEle, brushHandleEle);
                translate([x, y + bandBox.height + axisPadding + bandStrokeWidth + axisLinsStrokeWidth], axisEle, minmaxLabelEle); // TODO: consider brushBox
            }
            let titleX = mathMin(0, mathMin(getBoundingRect(minmaxLabelEle).left) - getBoundingRect(titleEle).left);
            translate([titleX, 0], titleEle);
        } else {
            if (orient === LEFT_ALIGN) {
                x = axisBox.width;
                y = titleBox.height + padding + -mathMin(0, minmaxLabelBox.y);

                translate([x + axisPadding + bandStrokeWidth, y], bandEle, brushEle, brushHandleEle);
                translate([x + bandBox.x - axisLinsStrokeWidth, y], axisEle, minmaxLabelEle);
            } else {
                y = titleBox.height + padding - mathMin(0, minmaxLabelBox.y); // TODO: check

                translate([x, y], bandEle, brushEle, brushHandleEle);
                translate([bandBox.width + bandBox.x + axisPadding + bandStrokeWidth + axisLinsStrokeWidth, y], axisEle);
                translate([bandBox.width + bandBox.x + bandStrokeWidth + axisLinsStrokeWidth, y], minmaxLabelEle);
            }
        }
    }
    getBrushOptions() {
        let { config, bound, area, format } = this.spec,
            { ranges } = config,
            { dataType, scale } = this.spec.scaleComponent,
            //
            align = getLegendAlignment(config),
            isHorizontal = align.layout === HORIZONTAL_ALIGN,
            dimension = legend_helpers.getColorBandDimension(config, bound, area),
            orient = legend_helpers.getOrient(config),
            reversed = legend_helpers.isReversedColorAxis(config),
            //
            _range = isHorizontal ? [0, dimension[0]] : [dimension[1], 0],
            range = reversed ? _range.reverse() : _range,
            //
            brushScale = getBrushScale(config, dataType, scale, range),
            initialExtent = getInitialExtent(ranges, config.filter.range, scale, brushScale, dataType),
            //
            brushOptions = {
                config,
                reversed,
                ranges,
                range,
                isHorizontal,
                initialExtent,
                align,
                format,
                brushScale,
                scale: brushScale,
                colorScale: scale,
                sizeScale: scale, // TODO: remove CXZA_SINGLEDATA_1 event
                datatype: dataType,
                dataType,
                type: isHorizontal ? 'x' : 'y',
                orient,
                dimension,
                scale_domain: scale.domain(),
                singleDomainContinuous: brushScale.singleDomainContinuous,
                encode: 'clr'
            };

        getHandleOptions(brushOptions);

        return brushOptions;
    }
}

function translate(xy, ...elements) {
    let _translate = `translate(${xy.join(' ')})`;
    each(elements, ele => ele.attr('transform', _translate));
}

function getEle(self, view) {
    let order = {
        LegendTitleSvg: 0,
        Band: 1,
        Axis: 2,
        MinmaxLabel: 3,
        Brush: 4,
        BrushHandle: 5
    };

    if (view === 'LegendBackgroundSvg') {
        return self.encode.children[0].getElement();
    }
    return self.encode.children[1].children[order[view]].getElement();
}

function getHandleOptions(options) {
    let { config, dimension, isHorizontal, ranges } = options,
        hoverOptions = config.itemHoverStyle || {}; //getHoverOptions(config, 'clr');

    if (hoverOptions.type) {
        let op = getDefaultSliderOptions(dimension, isHorizontal, ranges && ranges.length);
        config._itemHoverStyle = mergeJSON(op, hoverOptions, true);
    }
}

ColorLegendBand.encode = 'clr';
ColorLegendBand.view = 'band';
ColorLegendBand.domMode = 'svg';

export { ColorLegendBand };
Registry.setComponent('colorLegendBand', ColorLegendBand);
LegendRegistry.add(ColorLegendBand);
