// @zohocharts/zc/src/legend/color-legend/band/color-band-parser.js
//
import { BOTTOM_ALIGN, RIGHT_ALIGN } from '../../../core/align-constants';
import { LegendTitleSvgParser } from '../../title/legend-title-svg-parser';
import { _LegendBackgroundSvgParser } from '../../background/legend-background-svg-parser';
import { LegendContentWrapParser } from '../../size-legend/legend-content-wrap-parser';
import { BandParser } from './band-parser';
import { BandAxisParser } from './band-axis-parser';
import { MinmaxLabelParser } from '../../minmax-label/legend-minmax-label-parser';
import { BrushHandleParser } from './brush-handle-parser';
import { Mark } from '../../../mark/mark-utils';
import { d3_select } from '../../../core/d3-constants';
import { parseShortHandValue, isObjectPropDefined, getBoundingRect } from '../../../utils/zcutils';
import { doOverlap } from '../../../utils/geometry-utils';
import { getLegendAlignment } from '../../legend-utils';
import { legend_helpers } from '../../../core/main-zc-init';
import { getColorAxisOptions } from '../../legend-continuous';
import { getSliderSize } from '../../legend-minmaxlabels';
import { getLegendRanges } from '../../../axis/colorscale-helpers';
import { colorBandBrushOptions } from '../../legend-brush-options';
import { setDefaultOptions } from '../../../core/defaulttheme';

function ColorBandParser(obj) {
    let { titleSpec, viewSpec, axisSpec, backgroundSpec, minmaxLabelSpec, brushHandleSpec } = helper(obj.spec, obj.brushOptions); // TODO: implement axisEnd

    let groupMark = Mark('group', '.legendgroup'),
        // Background
        backgroundMark = _LegendBackgroundSvgParser(backgroundSpec),
        // Content wrap
        contentMark = LegendContentWrapParser(),
        // Title
        titleMark = LegendTitleSvgParser(titleSpec),
        // Band view
        bandMark = BandParser(viewSpec),
        //Brush handle
        brushHandleMark = BrushHandleParser(brushHandleSpec),
        // Axis
        axisMark = BandAxisParser(axisSpec),
        // Minmax labels
        minmaxLabelMark = MinmaxLabelParser(minmaxLabelSpec),
        // Brush
        brushMark = Mark('group', '.zc-band-brush');

    contentMark.children = [titleMark, bandMark, axisMark, minmaxLabelMark, brushMark, brushHandleMark];
    contentMark.onEnd = obj.align;
    contentMark.onStart = (ele) => hideHighlight(ele);
    groupMark.children = [backgroundMark, contentMark];

    return groupMark;
}

function helper(spec, brushOptions) {
    let { config, bound, area, scaleComponent, globalFont, callbackArgs } = spec;

    //Set default brush properties
    setDefaultOptions('filter', config, null, colorBandBrushOptions(config, bound, scaleComponent.dataType, area), true);

    let bandConfig = config.band || config.colorBand,
        axisConfig = getColorAxisOptions(config),
        minmaxLabelConfig = axisConfig.minmaxlabels,
        filterConfig = config.filter,
        { padding } = bandConfig,
        parsedPadding = parseShortHandValue(padding),
        //
        legendAlign = getLegendAlignment(config),
        isHorizontal = legendAlign.layout === 'horizontal',
        reversed = legend_helpers.isReversedColorAxis(config),
        dimension = legend_helpers.getColorBandDimension(config, bound, area),
        orient = legend_helpers.getOrient(config),
        sliderSize = filterConfig.enabled ? getSliderSize(filterConfig.slider, dimension, orient, legendAlign) : [0, 0],
        minmaxData = [];

    if (minmaxLabelConfig.show) {
        if (bandConfig.ranges && bandConfig.ranges.length) {
            let ranges = getLegendRanges(bandConfig, scaleComponent.dataType);
            minmaxData = [ranges[0][0], ranges[ranges.length - 1][ranges[ranges.length - 1].length - 1]];
        } else {
            let scale_domain = scaleComponent.scale.domain();
            minmaxData = [scale_domain[0], scale_domain[scale_domain.length - 1]];
        }
    }

    let availableBound = {
            maxWidth: bound.maxWidth - parsedPadding.left - parsedPadding.right,
            maxHeight: bound.maxHeight - parsedPadding.top - parsedPadding.bottom
        },
        minmaxLabelDim = dimension,
        axisOrient = axisConfig.orient || (isHorizontal ? BOTTOM_ALIGN : RIGHT_ALIGN);

    if (reversed) {
        minmaxData.reverse();
    }

    let titleSpec = {
            config: config.title,
            bound,
            margin: 5 // Remove (FEATURESEVENTS_HEATMAP_9)
        },
        viewSpec = {
            bound,
            config: bandConfig,
            dimension,
            patternSVG: spec.chartObj.getPatternSvg(), // TODO: remove
            defsPath: spec.chartObj.defsLocation, // TODO: remove
            reversed,
            isHorizontal,
            scale: scaleComponent.scale
        },
        axisSpec = {
            bound,
            scale: scaleComponent.scale,
            config: axisConfig,
            orient,
            dimension,
            format: spec.format,
            dataType: scaleComponent.dataType,
            meta: isObjectPropDefined(scaleComponent, 'schema.data.meta', true),
            bandConfig,
            isHorizontal,
            reversed,
            globalFont,
            callbackArgs,
            viewBound: { width: 0, height: 0 },
            availableBound: {
                maxWidth: bound.maxWidth - parsedPadding.left - parsedPadding.right - (!isHorizontal ? dimension[0] : 0), // TODO
                maxHeight: bound.maxHeight - parsedPadding.top - parsedPadding.bottom
            }
        },
        backgroundSpec = {
            config: config,
            margin: padding
        },
        minmaxLabelSpec = {
            format: spec.format, //d => getFormattedValue(formatConf, d, datatype, chartObj, 'minmaxlabels'), // TODO
            data: minmaxData,
            orient,
            dimension,
            sliderSize,
            config: minmaxLabelConfig,
            trimText: ele => {
                return;
                // let labelEle = ele.selectAll('text'),
                //     minmaxLabelAlign = axisConfig.minmaxlabels.align,
                //     _b = _getMinmaxLabelBound(
                //         availableBound,
                //         minmaxLabelDim,
                //         minmaxLabelAlign,
                //         isHorizontal,
                //         axisOrient,
                //         dimension,
                //         bound,
                //         sliderSize.map(_ => [_, _])
                //     ),
                //     textAnchor = minmaxLabelAlign === RIGHT_ALIGN ? 'start' : minmaxLabelAlign === LEFT_ALIGN ? 'end' : 'start',
                //     _labelParser = { anchor: textAnchor, margin: { left: 5, right: 5, bottom: 5, top: 5 } };
                //
                // handleTicklabels(ele, ele.selectAll('.nothing'), _b, { width: 0, height: 0 }, _labelParser);
                handleOverlapTickLabel(ele); // overlapping axis and minmax label
            }
        },
        brushHandleSpec = brushOptions;

    return { titleSpec, backgroundSpec, viewSpec, axisSpec, minmaxLabelSpec, brushHandleSpec };
}

// function _getMinmaxLabelBound(available, minmaxLabelDim, align, isHorizontal, orient, dimension, bound, sliderSize, i) {
//     let { maxWidth, maxHeight } = available;
//
//     if (isHorizontal && align === CENTER_ALIGN) {
//         if (i === 0) {
//             return { maxWidth, maxHeight, x: -(bound.maxWidth - dimension[0] - sliderSize[0][0] / 2 - sliderSize[1][0] / 2) / 2 };
//         }
//         return { maxWidth, maxHeight, x: (bound.maxWidth - dimension[0] - sliderSize[0][0] / 2 - sliderSize[1][0] / 2) / 2 };
//     }
//
//     return getMinmaxLabelBound(available, minmaxLabelDim, align, isHorizontal, orient, dimension, bound, sliderSize);
// }

function hideHighlight(ele) {
    // d3_select(mark.element.node().parentNode)
        d3_select(ele._parents[0].parentNode)
        .selectAll('g.highlight')
        .style('display', 'none');
}

// TODO: enhance
function handleOverlapTickLabel(minmaxLabel) {
    if (minmaxLabel.empty()) {
        return;
    }

    let mB = [];
    minmaxLabel.each(function() {
        mB.push(getRect(this));
    });

    let tickLabels = d3_select(minmaxLabel.node().parentNode.parentNode).selectAll('text.ticklabel');
    tickLabels.each(function() {
        let lB = getRect(this);

        if (doOverlap(lB, mB[0]) || doOverlap(lB, mB[1])) {
            d3_select(this).style('display', 'none');
        }
    });
}

function getRect(node) {
    let bb = getBoundingRect(node);
    return { x1: bb.left, y1: bb.top, x2: bb.right, y2: bb.bottom };
}

export { ColorBandParser };
