// @zohocharts/zc/src/gantt/connector-helpers.js
import { MarkerParser } from '../utils/marker-parser';
import { d3_line, d3_select, d3_curves } from '../core/d3-constants';
import { mathMin } from '../core/zc-constants';
import { DataProcessor } from '../core/data-processor';
import { defined, isArray, validateNullVal, pick } from '../utils/zcutils';
import { data } from '../mark/extender/data';
import { isWithinColorRange } from '../utils/colorutils';

export function getConnectorMark(options, type, order, prefix) {
    let _this = this,
        renderer = _this.oldRenderer;
    var rp = renderer.commonRendererProp().params;

    let dependencyConfig = this.builderC.commonPlotOptions.dependency,
        allDependencies = getDependencies(_this.data.data[0], dependencyConfig.sourceIndex, dependencyConfig.targetIndex);

    var symbolConf = dependencyConfig.marker,
        linkConf = dependencyConfig.link,
        linkPadding = defined(linkConf.padding) ? linkConf.padding : 1,
        symbolPadding = defined(symbolConf.padding) ? symbolConf.padding : linkPadding;
        symbolPadding = mathMin(linkPadding, symbolPadding);

    let axisrotated = _this.chart.axes.rotated, symbol;

    var ltype = linkConf.type;
    if (ltype == 'SS' || ltype == 'ES') {
        symbol = axisrotated ? 'triangle-right' : 'triangle-up';
    } else if (ltype == 'EE' || ltype == 'SE') {
        symbol = axisrotated ? 'triangle-left' : 'triangle-down';
    }

    symbolConf.symbol = symbolConf.symbol ? symbolConf.symbol : symbol;

    var arrowParser_inner = dependencyConfig.marker?.enabled ? getArrowParser(_this, type, symbolConf, order, "inner") : {};
    var arrowParser_outer = dependencyConfig.marker?.enabled ? getArrowParser(_this, type, symbolConf, order,  "outer") : {};

    let _styles = this.helpers.styles[type] ?? {};

    let baseG = {
        mark: 'g',
        selector: `.connector`,
        data: allDependencies ? allDependencies : [],
        children: [
            {
                mark: 'path',
                selector: '.link',
                data: function(d) {
                    return [d];
                },
                config: {
                    d: function(d, i) {
                        var details = _this.helpers.position[type].end(d, 'details'),
                            dependencyPathString = getDependencyPathString(details.type, details.end, details.start, linkPadding, symbolPadding, rp)(new Array(6));
                            dependencyPathString = linkConf.mode === 'curve' ? convertToCurvedPath(dependencyPathString) : dependencyPathString;
                        (dependencyPathString = _this.helpers.positionComponent[type].pathString(d, _this.builderC.datasetIndex, dependencyPathString));
                        return dependencyPathString
                    },
                    strokeWidth: _styles.strokeWidth,
                    dashStyle: _styles.strokeDasharray,
                    lineCap: _styles.lineCap
                },
                properties: {
                    styles: _this.helpers.styles.dependency,
                }
            },
            {
                mark: 'g',
                selector: '.arrow',
                data: function(d) {
                    return dependencyConfig.marker?.enabled ? [d] : [];
                },
                properties: {
                    attrs: {
                        transform: function(d) {
                            var temp,
                            details = _this.helpers.position['dependency'].end(d, 'details'),
                            x,
                            y;

                            let axisrotated = _this.chart.axes.rotated;

                            var type = details.type;
                            if (type == 'SS' || type == 'ES') {
                                x = axisrotated ? details.start.x - symbolPadding : details.start.x + details.start.width / 2;
                                y = axisrotated
                                    ? details.start.y + details.start.height / 2
                                    : details.start.y + details.start.height + symbolPadding;
                                temp = axisrotated ? 'triangle-right' : 'triangle-up';
                            } else if (type == 'EE' || type == 'SE') {
                                x = axisrotated ? details.start.x + details.start.width + symbolPadding : details.start.x + details.start.width / 2;
                                y = axisrotated ? details.start.y + details.start.height / 2 : details.start.y - symbolPadding;
                                temp = axisrotated ? 'triangle-left' : 'triangle-down';
                            }
                            return `translate(${x},${y})`;
                        }
                    },
                    styles: {
                        display: function(d){
                            if(_this.helpers.styles[type] && _this.helpers.styles[type].display){
                                return _this.helpers.styles[type].display(d);
                            }
                            return '';  
                        }
                    }
                },
                children: [{...arrowParser_outer}, {...arrowParser_inner}]
            } 
        ]
    };
    return baseG;
}


function getArrowParser(_this, type, options, order, prefix) {
    let options_cloned = { ...options };
    /*
    UPDATE STROKECOLOR & STROKEWIDTH IN OPTIONS
     */
    options_cloned.outerStrokeColor = d => {
        if (!isWithinColorRange(_this.chart.dataset.getClr(d), _this.chart)) {
            return 'transparent';
        } else {
            return options.outerStrokeColor;
        }
    };
    options_cloned.strokeWidth = options.strokeWidth || 0;

        /*
    HIDE INNER MARKER ON FOLLOWING CONDITIONS
     */
    let innerData;
    if (prefix == 'inner') {
        let hasImage = options.imageUrl != null && options.imageUrl !== '',
            innerColor = options.innerFillColor || options.color;
        innerData =
            hasImage || (options.innerStrokeWidth === 0 && (options.innerFillOpacity === 0 || innerColor === 'transparent')) ? [] : [1];
    }

    let _markerParser = MarkerParser({
        data: function(d) {
            return [_this.data.data[0]];
        },
        prefix: prefix || 'outer',
        selection: prefix ? `.${prefix}` : '.outer',
        config: {
            x: 0,
            y: 0,
            rotate: d => _this.helpers.position[type].end(d).rotate,
            ...options_cloned,
            shape: options.symbol,
            color: null,
            size: (d, i) => prefix == 'inner' ? _this.helpers.position[type].end(d, 'size', order, prefix) * 0.4 : _this.helpers.position[type].end(d, 'size', order, prefix) * 0.7
        },
        context: _this.chart.getPatternSvg(),
        defsPath: _this.chart.defsLocation
    });

    if (options.animation && options.animation.enabled) {
        delete _markerParser.config.strokeOpacity;
        _markerParser.config.enter = {
            strokeOpacity: 0
        };
        _markerParser.anime = {
            update: {
                duration: options.animation.duration,
                delay: options.animation.duration,
                strokeOpacity: options.outerStrokeOpacity
            }
        };
    }
    return _markerParser;
}


function getDependencies(source, dependencyIndex, searchIndex) {
    var allDependencies = [];
    source.forEach(function(dd) {
        var dependencies = [],
            currentDependency = dd[dependencyIndex];
        source.forEach(function(data) {
            if (data !== dd && defined(currentDependency) && isArray(currentDependency)) {
                if (currentDependency.indexOf(data[searchIndex]) !== -1) {
                    allDependencies.push({
                        source: dd,
                        target: data
                    });
                }
            }
        });
    });
    if (allDependencies.length) {
        return allDependencies;
    }
}


function isCrucialCurve(type, axisrotated, start, end, linkPadding) {
    var direct = false;
    switch (type) {
        case 'SS':
        case 'EE': {
            direct = true;
            break;
        }
        case 'SE': {
            direct = axisrotated ? start.x > end.x + end.width : start.y + start.height < end.y;
            break;
        }
        case 'ES': {
            direct = axisrotated ? start.x + start.width < end.x - linkPadding * 2 : start.y > end.y + end.height;
        }
    }
    return !direct;
}

function getDependencyPathString(type, startPosition, endPosition, padding, markerPadding, rp) {
    type = type.toUpperCase();

    var chartObj = rp.chartObj,
        axisrotated = rp.axisrotated,
        xAxisReverseed = chartObj.systemConf.chart.axes.xaxis.reversed,
        x = null,
        y = null,
        pos = null,
        crucialCurve = isCrucialCurve(type, axisrotated, startPosition, endPosition, padding);

    return d3_line()
        .x(function(d, j) {
            switch (j) {
                case 0:
                case 1:
                case 2: {
                    pos = startPosition;
                    if (j == 0) {
                        if (type == 'SS' || type == 'SE') {
                            x = axisrotated ? pos.x : pos.x + pos.width / 2;
                        } else if (type == 'EE' || type == 'ES') {
                            x = axisrotated ? pos.x + pos.width : pos.x + pos.width / 2;
                        }
                    } else if (j == 1) {
                        if (type == 'SS' || type == 'SE') {
                            x = axisrotated ? pos.x - padding : x;
                        } else if (type == 'EE' || type == 'ES') {
                            x = axisrotated ? pos.x + pos.width + padding : x;
                        }
                    } else {
                        var dist;
                        if (type == 'SS') {
                            dist = startPosition.x - endPosition.x;
                            dist = dist < 0 ? 0 : dist;
                            x = axisrotated ? pos.x - dist - padding : x;
                        } else if (type == 'EE') {
                            dist = startPosition.x + startPosition.width - (endPosition.x + endPosition.width);
                            dist = dist > 0 ? 0 : dist;
                            x = axisrotated ? pos.x + pos.width + padding - dist : x;
                        } else if (type == 'SE') {
                            if (axisrotated) {
                                x = pos.x - padding;
                                y = crucialCurve ? pos.y + pos.height : endPosition.y;
                            } else {
                                x = xAxisReverseed ? x - pos.width : x + pos.width;
                            }
                        } else if (type == 'ES') {
                            if (axisrotated) {
                                x = pos.x + padding + pos.width;
                                y = crucialCurve ? pos.y + pos.height : endPosition.y;
                            } else {
                                x = xAxisReverseed ? x - pos.width : x + pos.width;
                            }
                        }
                    }
                    break;
                }
                default: {
                    if (axisrotated) {
                        y = null;
                    }
                    pos = endPosition;
                    if (j == 3) {
                        if (axisrotated && crucialCurve) {
                            y = startPosition.y + startPosition.height;
                        }
                        if (type == 'SS' || type == 'EE') {
                            x = axisrotated ? x : pos.x + pos.width / 2;
                        } else if (type == 'SE') {
                            x = axisrotated ? pos.x + pos.width + padding : x;
                        } else if (type == 'ES') {
                            x = axisrotated ? pos.x - padding : x;
                        }
                    } else if (j == 4) {
                        if (type == 'SE') {
                            x = axisrotated ? pos.x + pos.width + padding : pos.x + pos.width / 2;
                        } else if (type == 'ES') {
                            x = axisrotated ? x : pos.x + pos.width / 2;
                        }
                    } else {
                        if (type == 'SS' || type == 'ES') {
                            x = axisrotated ? pos.x - markerPadding : x;
                        } else if (type == 'EE' || type == 'SE') {
                            x = axisrotated ? pos.x + pos.width + markerPadding : x;
                        }
                    }
                }
            }
            return x;
        })
        .y(function(d, j) {
            if (axisrotated) {
                return (y || pos.y) + pos.height / 2;
            }

            if (j == 0) {
                if (type == 'SS' || type == 'SE') {
                    y = startPosition.y + startPosition.height;
                } else if (type == 'EE' || type == 'ES') {
                    y = startPosition.y;
                }
            } else if (j == 1) {
                if (type == 'SS' || type == 'SE') {
                    y = y + padding;
                } else if (type == 'EE' || type == 'ES') {
                    y = startPosition.y - padding;
                }
            } else if (j == 2) {
                if (type == 'SS') {
                    y =
                        startPosition.y + startPosition.height < endPosition.y + endPosition.height
                            ? endPosition.y + endPosition.height + padding
                            : y;
                } else if (type == 'EE') {
                    y = startPosition.y > endPosition.y ? endPosition.y - padding : y;
                }
            } else if (j == 3) {
                if (type == 'SE') {
                    y = crucialCurve ? endPosition.y - padding : y;
                } else if (type == 'ES') {
                    y = crucialCurve ? endPosition.y + endPosition.height + padding : y;
                }
            } else if (j == 4) {
                y = y + 0;
            } else {
                if (type == 'SS' || type == 'ES') {
                    y = endPosition.y + endPosition.height + markerPadding;
                } else if (type == 'EE' || type == 'SE') {
                    y = endPosition.y - markerPadding;
                }
            }
            return y;
        })
        .curve(d3_curves('linear'));
}


function convertToCurvedPath(pathString, radius = 20) {
    // Split the path string into commands
    const commands = pathString.match(/[A-Z][^A-Z]*/g) || [];
    const points = commands.map(cmd => {
      const type = cmd[0];
      const coords = cmd.slice(1).trim().split(',').map(Number);
      return { type, x: coords[0], y: coords[1] };
    });
  
    // Remove the last Z if it exists
    const hasClosePath = points[points.length - 1].type === 'Z';
    if (hasClosePath) points.pop();
  
    let curvedPath = '';
    const numPoints = points.length;
  
    // Helper function to calculate angle between vectors
    function getAngleBetweenVectors(dx1, dy1, dx2, dy2) {
      const dot = dx1 * dx2 + dy1 * dy2;
      const mag1 = Math.sqrt(dx1 * dx1 + dy1 * dy1);
      const mag2 = Math.sqrt(dx2 * dx2 + dy2 * dy2);
      
      // Check for zero magnitudes
      if (mag1 === 0 || mag2 === 0) return 180;
      
      const cosTheta = dot / (mag1 * mag2);
      // Ensure the value is within valid range for acos
      const safeValue = Math.min(1, Math.max(-1, cosTheta));
      return Math.acos(safeValue) * (180 / Math.PI);
    }
  
    for (let i = 0; i < points.length; i++) {
      const point = points[i];
      
      if (i === 0) {
        curvedPath += `M${point.x},${point.y}`;
        continue;
      }
  
      const prev = points[i - 1];
        
      // If this is the last point, just draw a straight line to it
      if (i === numPoints - 1) {
        curvedPath += ` L${point.x},${point.y}`;
        continue;
      }
  
      const next = points[i + 1];
  
      // Calculate vectors
      const dx1 = point.x - prev.x;
      const dy1 = point.y - prev.y;
      const d1 = Math.sqrt(dx1 * dx1 + dy1 * dy1);
      
      const dx2 = next.x - point.x;
      const dy2 = next.y - point.y;
      const d2 = Math.sqrt(dx2 * dx2 + dy2 * dy2);
  
      // If either segment is too short, draw a straight line
      if (d1 < 0.1 || d2 < 0.1) {
        curvedPath += ` L${point.x},${point.y}`;
        continue;
      }
  
      // Calculate angle between segments
      const angle = getAngleBetweenVectors(dx1, dy1, dx2, dy2);
  
      // Skip curves for near-straight lines or final segments
      if (angle > 165) {
        curvedPath += ` L${point.x},${point.y}`;
        continue;
      }
  
      // Use a smaller radius if the segment is too short
      const useRadius = Math.min(radius, d1 / 3, d2 / 3);
  
      // Calculate points before and after the curve
      const beforeX = point.x - (dx1 * useRadius / d1);
      const beforeY = point.y - (dy1 * useRadius / d1);
  
      // Draw line to the point before curve
      curvedPath += ` L${beforeX},${beforeY}`;
  
      // Add the curve
      const afterX = point.x + (dx2 * useRadius / d2);
      const afterY = point.y + (dy2 * useRadius / d2);
      
      // Verify that all numbers are valid before adding to path
      if (!isNaN(beforeX) && !isNaN(beforeY) && !isNaN(afterX) && !isNaN(afterY)) {
        curvedPath += ` Q${point.x},${point.y} ${afterX},${afterY}`;
      } else {
        // Fallback to straight line if calculations produce invalid numbers
        curvedPath += ` L${point.x},${point.y}`;
      }
    }
  
    // Close the path if original was closed
    if (hasClosePath) curvedPath += ' Z';
  
    return curvedPath;
  }
