// @zohocharts/zc/src/gantt/gantt-marker-encode.js
import { getBarMarkerSize } from '../bar/bar-marker-helpers';
import { defined, pick } from '../utils/zcutils';
import { whiskerSize } from '../boxplot/whiskers';
import { isMultiColoring, getStrokeColor, getFillColor } from '../utils/colorutils';
import { effects } from "../effects/zceffects";

export let gantt_marker_position = function() {
    let position_helpers = {};

    const KEY = 'marker';

    if (this.builderC.commonPlotOptions[KEY] && this.builderC.commonPlotOptions[KEY].enabled) {
        let _this = this,
            builderC = this.builderC,
            markerOptions = builderC.commonPlotOptions[KEY];

        this.helpers.position[KEY] = {
            start: (d, encode) => {},
            end: (d, encode, order, prefix) => {

                let pos = order === 0 ? "start" : "end";

                if (encode == 'size') {
                    if(prefix == 'inner') {
                        return [markerOptions[pos].size * 10] * 0.6;
                    }
                    return [markerOptions[pos].size * 10];
                }

                var position = getMarkerPosition(_this, d, 0, null, _this.oldRenderer),
                    x = 0,
                    y = 0;

                if (order === 0) {
                    x = position.x + position.width / 2;
                    y = position.y - (markerOptions.start.size * 10) / 2;
                } else {
                    x = position.x + position.width / 2;
                    y = position.y + position.height + (markerOptions.end.size * 10) / 2;
                }

                ({x, y} = this.helpers.positionComponent[KEY].coordinates(d, this.builderC.datasetIndex, {x, y}));

                return {
                    x,
                    y,
                    rotate: this.chart.rotated ? 90 : 0
                };
            }
        }
    }

    return position_helpers;
};

export let gantt_marker_styles = function() {

    const KEY = 'marker';
    
    this.helpers.positionComponent[KEY] = {};
    this.helpers.positionComponent[KEY] = {
        coordinates: (d, i, defaultCoordinates = {x, y}) => {
            return defaultCoordinates
        }
    }

    let options = this.builderC.commonPlotOptions[KEY].start;

    if(this.builderC.commonPlotOptions[KEY] && this.builderC.commonPlotOptions[KEY].enabled) {

        this.helpers.styles[KEY] = {
            fill: options.color,
            stroke: options.color,
            strokeWidth: pick(options.strokeWidth, 1),
            strokeDasharray: options.dashStyle,
            strokeLinecap: options.lineCap,
            display: (d, i) => {
                return '';
            }
        };
    };
};

var getMarkerPosition = function(_this, d, i, j, renderer) {
    var width,
        height,
        x,
        y,
        cp = renderer.commonOption;
    x = _this.chart.dataObject.getTransRectValue(cp.axisrotated ? 'y' : 'x', d, i, null, cp);
    y = _this.chart.dataObject.getTransRectValue(cp.axisrotated ? 'x' : 'y', d, i, null, cp);
    width = _this.chart.dataObject.getTransRectValue(cp.axisrotated ? 'height' : 'width', d, i, null, cp);
    height = _this.chart.dataObject.getTransRectValue(cp.axisrotated ? 'width' : 'height', d, i, null, cp);
    return {
        x: x,
        y: y,
        width: width,
        height: height
    };
};
