// @zohocharts/zc/src/gantt/gantt-highlight.js
import { events, plot, svg_plot_renderer, tempCanvasContext } from '../core/main-zc-init';
import { mathMin, mathMax, mathAbs } from '../core/zc-constants';
import { _super, inherit, mergeJSON } from '../utils/zcutils';
import { BarHighlighter } from '../bar/bar-highlight';
import { patternFillEffect } from '../events/visualEffects/highlight/pattern-fill';
import { appendEle, getCorrectedRectParams, roundedRect } from '../utils/zcsvgutils';
import { NULL } from '../core/zc-constants';
import { HIGHLIGHT_CONTEXT } from '../events/visualEffects/highlight/highlight-context';
import { addSVGmarker } from '../utils/svg-symbols';

export function GanttHighlighter(chart, renderer) {
    this.chart = chart;
    this.renderer = renderer;
}

inherit(GanttHighlighter, BarHighlighter);

var proto = GanttHighlighter.prototype;
proto._defaultH = function() {
    return [
        {
            shape: 'rect',
            path: _super(this).path,
            data: _super(this).data,
            context: _super(this).renderingContext,
            translate: _super(this).translateContext,
            styles: patternFillEffect
        },
        {
            shape: 'rect',
            path: this.levelMarker,
            context: this.renderingContext,
            translate: _super(this).translateContext,
            styles: point => {
                return this.getStylesForLevelMarker.call(this, point);
            }
        },
        {
            shape: 'marker',
            path: this.markerPath,
            context: this.markerContext,
            translate: _super(this).translateContext,
            type: 'start'
        },
        {
            shape: 'marker',
            path: this.markerPath,
            context: this.markerContext,
            translate: _super(this).translateContext,
            type: 'end'
        }
    ];
};

proto.levelMarker = function(point, model) {
    var chart = this.chart,
        renderer = this.renderer,
        rp = renderer.commonRendererProp().params,
        levelMarker = rp.levelMarker || {};

    let p = svg_plot_renderer.ganttHelpers.getLevelMarkerParams(point, point.subSeriesIndex, point.itemIndex, this.renderer, this.renderer.commonOption);

    var rect = roundedRect(
        p.x,
        p.y,
        p.width,
        p.height,
        this.chart.dataObject.getTransRectValue('radius', null, null, null, 0),
        rp.yaxisreversed,
        this.chart.axes.rotated
    );

    var context = model.context.call(this),
        pathContext = appendEle(context, 'path', levelMarker.enabled ? [1] : [], '.', 'id', `levelMarkerpath`);
    context.style('display', '');
    pathContext
        .style('display', '')
        .attr('d', rect)
        .attr('transform', `translate(${this.chart.plotarea.left},${this.chart.plotarea.top})`);
    return pathContext;
};

proto.markerPosition = function(point, model) {
    let renderer = this.renderer,
        align = model.type,
        isHorizontal = this.chart.axes.rotated,
        orient = getMarkerOrient(align, isHorizontal),
        x,
        y;
    switch (orient) {
        case 'left':
            x = point._x;
            y = point._y + point._height / 2;
            break;
        case 'right':
            x = point._x + point._width;
            y = point._y + point._height / 2;
            break;
        case 'top':
            x = point._x + point._width / 2;
            y = point._y;
            break;
        case 'bottom':
            x = point._x + point._width / 2;
            y = point._y + point._height;
            break;
    }

    return {
        x: x,
        y: y
    };
};

proto.markerPath = function(point, model) {
    var _options = this.markerOptions(point, model.type),
        _hoveroptions = _options.options,
        p1 = this.markerPosition(point, model);
    if (!_options.enabled) {
        return;
    }
    var context = model.context.call(this),
        translate = model.translate.call(this),
        pathContext = appendEle(context, 'g', [1], '.', 'class', `yrangePos${model.type}`);
    addSVGmarker(
        pathContext,
        null,
        _hoveroptions,
        {
            color: _options.color,
            symbolSize: {
                inner: _options.size.inner,
                outer: _options.size.outer
            },
            transform: [p1.x, p1.y],
            imageSize: 15
        },
        this.chart
    );
    context.attr('transform', `translate(${translate[0]},${translate[1]})`);
    return pathContext;
};

proto.markerOptions = function(point, type) {
    var chart = this.chart,
        renderer = this.renderer,
        rp = renderer.commonRendererProp().params,
        cpo = rp.commonPlotOptions.hoveroptions[type],
        spo = this.chart.seriesdata[point.seriesIndex].plotoptions,
        spo = spo && spo.hoveroptions ? spo.hoveroptions[type] : {},
        _options = mergeJSON(cpo, spo, true),
        cpo_marker = rp.commonPlotOptions.marker[type],
        spo_marker = spo && spo.marker ? spo.marker[type] : {},
        _options_marker = mergeJSON(cpo_marker, spo_marker, true),
        _options = mergeJSON(_options_marker, _options, true);

    return {
        enabled: rp.commonPlotOptions.marker && rp.commonPlotOptions.marker.enabled,
        options: _options,
        symbol: _options.symbol,
        _size: _options.size,
        color: this.options(point).color,
        size: {
            outer: _options.size * 2,
            inner: _options.size * 1
        }
    };
};
proto.markerContext = function() {
    return HIGHLIGHT_CONTEXT.get('scatter').call(this, this.chart);
};
proto.getStylesForLevelMarker = function(point) {
    var renderer = this.renderer,
        rp = renderer.commonRendererProp().params,
        levelMarker = rp.levelMarker || {},
        options = this.options(point);

    return {
        fill: levelMarker.color || options.color
    };
};

/*
Consider four sided rectangle
Need to place the marker at any of the centre of four sides
For vertical rectangle - > marker should be placed on top and bottom
For horizontal rectangle - > marker should be placed on left and right
 */
function getMarkerOrient(align, isHorizontal) {
    if (isHorizontal) {
        return align == 'start' ? 'left' : 'right';
    } else {
        return align == 'start' ? 'top' : 'bottom';
    }
}
