// @zohocharts/zc/src/gantt/ganttUtils.js
//ganttUtils.js
import { svg_plot_renderer, plot, chartcategory, charttype } from '../core/main-zc-init';
import { defined, isArray, validateNullVal, pick } from '../utils/zcutils';
import { getDashArray, getLineCap } from '../utils/dashstyles';
import { mathAbs, mathMin } from '../core/zc-constants';
import { appendEle } from '../utils/zcsvgutils';
import { d3_line, d3_select, d3_curves } from '../core/d3-constants';
import { addSVGmarker } from '../utils/svg-symbols';
import { whiskerSize } from '../boxplot/whiskers';
import { isWithinColorRange } from '../utils/colorutils';

svg_plot_renderer.ganttHelpers = function() {};
plot.render.ganttHelpers = function() {};

svg_plot_renderer.ganttHelpers.showDependencies = function(renderer, data, series, index) {
    var rp = renderer.commonRendererProp().params,
        cp = renderer.commonOption,
        chartObj = rp.chartObj,
        plotarea = chartObj.plotarea,
        dataset = rp.dataset,
        axisrotated = rp.axisrotated,
        commonPlotOptions = rp.commonPlotOptions,
        yaxiscolumnorder = cp.yaxiscolumnorder,
        color = rp.color;

    var dependencyConfig = commonPlotOptions.dependency,
        showDependency = dependencyConfig.enabled === true;

    var dependencies = null;
    var dependenceSeriesNode = series.selectAll('g.dependencySeries').data(showDependency ? [data] : []);

    var dependenceSeries = dependenceSeriesNode
        .enter()
        .insert('g', ':first-child')
        .attr('class', 'dependencySeries')
        .merge(dependenceSeriesNode);
    // dependenceSeries.attr('class', 'dependencySeries');
    dependenceSeries.exit().remove();
    if (showDependency && defined(dependencyConfig.sourceIndex) && defined(dependencyConfig.targetIndex)) {
        dependencies = getDependencies(data, +dependencyConfig.sourceIndex, +dependencyConfig.targetIndex);
    }
    if (dependencies) {
        var symbolConf = dependencyConfig.marker,
            linkConf = dependencyConfig.link,
            linkPadding = defined(linkConf.padding) ? linkConf.padding : 1,
            symbolPadding = defined(symbolConf.padding) ? symbolConf.padding : linkPadding;
        symbolPadding = mathMin(linkPadding, symbolPadding);
        // dependency links

        var linkGroup = appendEle(dependenceSeries, 'g', [1], '.', 'class', 'links');
        var arrowGroup = appendEle(dependenceSeries, 'g', [1], '.', 'class', 'symbol');

        // links
        var path = appendEle(linkGroup, 'path', dependencies, '.', 'class', 'link');
        path.attr('d', function(d, j) {
            var details = getDependencyDetails(d, dependencyConfig, renderer, index);
            return getDependencyPathString(details.type, details.end, details.start, linkPadding, symbolPadding, rp)(new Array(6));
        })
            .attr('stroke', function(d) {
                var startTime = dataset.getY(d.source, yaxiscolumnorder)[0];
                var endTime = dataset.getY(d.target, yaxiscolumnorder)[1];
                var error = startTime < endTime;
                return error ? linkConf.errorStyle : linkConf.strokeColor;
            })
            .attr('stroke-width', linkConf.strokeWidth)
            .attr('fill', 'none');

        // symble

        var markerEnabled = dependencyConfig.marker.enabled === true;
        var arrows = appendEle(arrowGroup, 'g', markerEnabled ? dependencies : [], '.', 'class', 'arrow');

        if (dependencyConfig.marker.enabled) {
            arrows.each(function(d, j) {
                var symbolType = symbolConf.symbol,
                    temp;
                var markerID,
                    mSize = {
                        inner: 0.4,
                        outer: 0.7
                    };
                var self = d3_select(this);
                self.attr('transform', function(d) {
                    var details = getDependencyDetails(d, dependencyConfig, renderer, index),
                        x,
                        y;
                    var type = details.type;
                    if (type == 'SS' || type == 'ES') {
                        x = axisrotated ? details.start.x - symbolPadding : details.start.x + details.start.width / 2;
                        y = axisrotated
                            ? details.start.y + details.start.height / 2
                            : details.start.y + details.start.height + symbolPadding;
                        temp = axisrotated ? 'triangle-right' : 'triangle-up';
                    } else if (type == 'EE' || type == 'SE') {
                        x = axisrotated ? details.start.x + details.start.width + symbolPadding : details.start.x + details.start.width / 2;
                        y = axisrotated ? details.start.y + details.start.height / 2 : details.start.y - symbolPadding;
                        temp = axisrotated ? 'triangle-left' : 'triangle-down';
                    }
                    return `translate(${x},${y})`;
                });
                addSVGmarker(
                    self,
                    null,
                    symbolConf,
                    {
                        symbolSize: mSize,
                        color: null,
                        symbol: temp
                    },
                    chartObj
                );
            });
        }
    }
};

function getDependencyDetails(d, dependencyConfig, renderer, i) {
    return {
        start: getPos(d.source, renderer, i),
        end: getPos(d.target, renderer, i),
        type: dependencyConfig.link.type.toUpperCase() || 'ES'
    };
}

function getDependencyPathString(type, startPosition, endPosition, padding, markerPadding, rp) {
    type = type.toUpperCase();

    var chartObj = rp.chartObj,
        axisrotated = rp.axisrotated,
        xAxisReverseed = chartObj.systemConf.chart.axes.xaxis.reversed,
        x = null,
        y = null,
        pos = null,
        crucialCurve = isCrucialCurve(type, axisrotated, startPosition, endPosition, padding);

    return d3_line()
        .x(function(d, j) {
            switch (j) {
                case 0:
                case 1:
                case 2: {
                    pos = startPosition;
                    if (j == 0) {
                        if (type == 'SS' || type == 'SE') {
                            x = axisrotated ? pos.x : pos.x + pos.width / 2;
                        } else if (type == 'EE' || type == 'ES') {
                            x = axisrotated ? pos.x + pos.width : pos.x + pos.width / 2;
                        }
                    } else if (j == 1) {
                        if (type == 'SS' || type == 'SE') {
                            x = axisrotated ? pos.x - padding : x;
                        } else if (type == 'EE' || type == 'ES') {
                            x = axisrotated ? pos.x + pos.width + padding : x;
                        }
                    } else {
                        var dist;
                        if (type == 'SS') {
                            dist = startPosition.x - endPosition.x;
                            dist = dist < 0 ? 0 : dist;
                            x = axisrotated ? pos.x - dist - padding : x;
                        } else if (type == 'EE') {
                            dist = startPosition.x + startPosition.width - (endPosition.x + endPosition.width);
                            dist = dist > 0 ? 0 : dist;
                            x = axisrotated ? pos.x + pos.width + padding - dist : x;
                        } else if (type == 'SE') {
                            if (axisrotated) {
                                x = pos.x - padding;
                                y = crucialCurve ? pos.y + pos.height : endPosition.y;
                            } else {
                                x = xAxisReverseed ? x - pos.width : x + pos.width;
                            }
                        } else if (type == 'ES') {
                            if (axisrotated) {
                                x = pos.x + padding + pos.width;
                                y = crucialCurve ? pos.y + pos.height : endPosition.y;
                            } else {
                                x = xAxisReverseed ? x - pos.width : x + pos.width;
                            }
                        }
                    }
                    break;
                }
                default: {
                    if (axisrotated) {
                        y = null;
                    }
                    pos = endPosition;
                    if (j == 3) {
                        if (axisrotated && crucialCurve) {
                            y = startPosition.y + startPosition.height;
                        }
                        if (type == 'SS' || type == 'EE') {
                            x = axisrotated ? x : pos.x + pos.width / 2;
                        } else if (type == 'SE') {
                            x = axisrotated ? pos.x + pos.width + padding : x;
                        } else if (type == 'ES') {
                            x = axisrotated ? pos.x - padding : x;
                        }
                    } else if (j == 4) {
                        if (type == 'SE') {
                            x = axisrotated ? pos.x + pos.width + padding : pos.x + pos.width / 2;
                        } else if (type == 'ES') {
                            x = axisrotated ? x : pos.x + pos.width / 2;
                        }
                    } else {
                        if (type == 'SS' || type == 'ES') {
                            x = axisrotated ? pos.x - markerPadding : x;
                        } else if (type == 'EE' || type == 'SE') {
                            x = axisrotated ? pos.x + pos.width + markerPadding : x;
                        }
                    }
                }
            }
            return x;
        })
        .y(function(d, j) {
            if (axisrotated) {
                return (y || pos.y) + pos.height / 2;
            }

            if (j == 0) {
                if (type == 'SS' || type == 'SE') {
                    y = startPosition.y + startPosition.height;
                } else if (type == 'EE' || type == 'ES') {
                    y = startPosition.y;
                }
            } else if (j == 1) {
                if (type == 'SS' || type == 'SE') {
                    y = y + padding;
                } else if (type == 'EE' || type == 'ES') {
                    y = startPosition.y - padding;
                }
            } else if (j == 2) {
                if (type == 'SS') {
                    y =
                        startPosition.y + startPosition.height < endPosition.y + endPosition.height
                            ? endPosition.y + endPosition.height + padding
                            : y;
                } else if (type == 'EE') {
                    y = startPosition.y > endPosition.y ? endPosition.y - padding : y;
                }
            } else if (j == 3) {
                if (type == 'SE') {
                    y = crucialCurve ? endPosition.y - padding : y;
                } else if (type == 'ES') {
                    y = crucialCurve ? endPosition.y + endPosition.height + padding : y;
                }
            } else if (j == 4) {
                y = y + 0;
            } else {
                if (type == 'SS' || type == 'ES') {
                    y = endPosition.y + endPosition.height + markerPadding;
                } else if (type == 'EE' || type == 'SE') {
                    y = endPosition.y - markerPadding;
                }
            }
            return y;
        })
        .curve(d3_curves('linear'));
}

function isCrucialCurve(type, axisrotated, start, end, linkPadding) {
    var direct = false;
    switch (type) {
        case 'SS':
        case 'EE': {
            direct = true;
            break;
        }
        case 'SE': {
            direct = axisrotated ? start.x > end.x + end.width : start.y + start.height < end.y;
            break;
        }
        case 'ES': {
            direct = axisrotated ? start.x + start.width < end.x - linkPadding * 2 : start.y > end.y + end.height;
        }
    }
    return !direct;
}

function getDependencies(source, dependencyIndex, searchIndex) {
    var allDependencies = [];
    source.forEach(function(dd) {
        var dependencies = [],
            currentDependency = dd[dependencyIndex];
        source.forEach(function(data) {
            if (data !== dd && defined(currentDependency) && isArray(currentDependency)) {
                if (currentDependency.indexOf(data[searchIndex]) !== -1) {
                    allDependencies.push({
                        source: dd,
                        target: data
                    });
                }
            }
        });
    });
    if (allDependencies.length) {
        return allDependencies;
    }
}

function getPos(d, renderer, i) {
    var width,
        height,
        x,
        y,
        cp = renderer.commonOption;
    x = renderer.getParamValue('x', d, i, null, cp);
    y = renderer.getParamValue('y', d, i, null, cp);
    width = renderer.getParamValue('width', d, i, null, cp);
    height = renderer.getParamValue('height', d, i, null, cp);
    return {
        x: x,
        y: y,
        width: width,
        height: height
    };
}

svg_plot_renderer.ganttHelpers.addMarker = function(series, data, renderer, index, markerOption, multiColoring, id) {
    var rp = renderer.commonRendererProp().params,
        cp = renderer.commonOption,
        chartObj = rp.chartObj,
        plotarea = chartObj.plotarea,
        dataset = rp.dataset,
        axisrotated = rp.axisrotated,
        color = rp.color,
        levelMarkerWidth = cp.levelMarkerWidth,
        cType = cp.charttype,
        isBullet = cType == 'bullet',
        isGantt = cType == 'gantt',
        isBoxplot = cType == 'boxplot',
        isWhiskerPlot = chartcategory.whiskerPlot.has(charttype[cType]),
        isWhisker = markerOption.symbol == 'whiskers' || markerOption.symbol == 'horizontal-whisker',
        isWhiskerOnEdges = isWhisker && rp.whiskers.showOn == 'edges',
        bandwidth = isBullet ? rp.xscale.bandwidth() : levelMarkerWidth,
        bandwidth1 = renderer.bandwidth1(),
        customMarkerSize = markerOption.size,
        iscustomMarkerSize = defined(customMarkerSize),
        size = iscustomMarkerSize ? customMarkerSize : levelMarkerWidth - 2 * validateNullVal(markerOption.padding, cp.padding / 2),
        target = appendEle(series, 'g', [data], '.', 'class', `markers${id}`, null, function(d) {
            return [dataset.getX(d)];
        });

    if (iscustomMarkerSize && typeof size === 'string' && size.indexOf('%') > -1) {
        //if user gives size as percent, '75%'
        size = ((parseFloat(size) / 100) * bandwidth) / 10;
    }

    // var markerWidth = ((markerOption.strokeWidth == null || markerOption.strokeWidth === "") ? 1 : JSON.parse(markerOption.strokeWidth));
    var markerWidth = pick(markerOption.strokeWidth, 1);

    // target.enter().append("g")
    // target.exit().remove();
    target.attrs({
        class: `markers${id}`,
        transform: `translate(${plotarea.left},${plotarea.top})`,
        'stroke-width': markerWidth,
        'stroke-dasharray': getDashArray(markerOption.dashStyle, validateNullVal(markerOption.outerStrokeWidth, 1)),
        'stroke-linecap': getLineCap(markerOption.lineCap, markerOption.dashStyle),
        fill: markerOption.color,
        stroke: markerOption.color
    });
    // isWithinColorRange(dataset.getClr(d), chartObj) ? dataset.getX(d) : null
    var marker = appendEle(target, 'g', data, null, null, null, null, function(d) {
        return dataset.getX(d);
    });
    marker.enter().append('g');
    var markerID,
        mSize =
            markerOption.symbol === 'line'
                ? {
                      inner: 0,
                      outer: 1
                  }
                : {
                      inner: 0.6,
                      outer: 1
                  };

    marker.each(function(d, j) {
        var m = d3_select(this);
        var clonedMarkerOptions;
        var clr = multiColoring ? chartObj.seriesColor[j % chartObj.seriesColor.length] : color;
        m.attrs({
            transform: function(d, k) {
                var x = 0,
                    y = 0;
                if (isGantt) {
                    var position = svg_plot_renderer.ganttHelpers.getMarkerPosition(d, index, k, renderer);
                    if (id === 0) {
                        x = position.x + position.width / 2;
                        y = position.y - (markerOption.size * 10) / 2;
                    } else {
                        x = position.x + position.width / 2;
                        y = position.y + position.height + (markerOption.size * 10) / 2;
                    }
                    return `translate(${axisrotated ? y : x},${axisrotated ? x : y})`;
                } else {
                    x = bandwidth1(dataset.getX(d)) + bandwidth / 2;
                    x = isWhiskerPlot ? x + cp.bandwidth2(cp.barDatasetIndex) + cp.padding : x; //For boxplot median and whiskers need to add bandwidth2
                    y = isWhisker
                        ? isWhiskerOnEdges
                            ? rp.yscale(d[markerOption.dataindex][id])
                            : (rp.yscale(d[markerOption.dataindex][0]) + rp.yscale(d[markerOption.dataindex][1])) / 2
                        : rp.yscale(d[markerOption.dataindex]);
                    return d[markerOption.dataindex] == null
                        ? null
                        : `translate(${axisrotated ? y : x},${axisrotated ? x : y}) rotate(${axisrotated ? 90 : 0})`;
                }
            }
        }).style('display', function(d) {
            return d[markerOption.dataindex] == null && isBullet ? 'none' : null;
        });

        markerOption.size = iscustomMarkerSize ? (isBullet ? mathMin(size, bandwidth / 10) : size) : size / 10;
        var customMarkers, wSize;
        if (isWhiskerPlot && isWhisker) {
            if (isWhiskerOnEdges) {
                wSize = whiskerSize(d, rp.yscale, markerOption.dataindex, dataset.yidx(cp.yaxiscolumnorder), id);
                wSize = axisrotated ? -wSize : wSize;
                customMarkers = {
                    symbol: 'vertical-line-z',
                    size: wSize
                };
            } else {
                wSize = whiskerSize(d, rp.yscale, null, null, null, markerOption.dataindex);
            }
        }
        if (!isWithinColorRange(dataset.getClr(d), chartObj)) {
            clonedMarkerOptions = { ...markerOption };
            clonedMarkerOptions.outerStrokeColor = 'transparent';
        }
        addSVGmarker(
            m,
            null,
            clonedMarkerOptions || markerOption,
            {
                symbolSize: mSize,
                color: isGantt ? clr : null,
                custom: customMarkers,
                size2d: wSize
            },
            chartObj
        );
    });
    markerOption.size = customMarkerSize; //reset with actual value

    if (isGantt) {
        renderer.markerProperties = {
            subseriesIndex: index
        };
    }
};

svg_plot_renderer.ganttHelpers.getMarkerPosition = function(d, i, j, renderer) {
    var width,
        height,
        x,
        y,
        cp = renderer.commonOption;
    x = renderer.getParamValue(cp.axisrotated ? 'y' : 'x', d, i, j, cp);
    y = renderer.getParamValue(cp.axisrotated ? 'x' : 'y', d, i, j, cp);
    width = renderer.getParamValue(cp.axisrotated ? 'height' : 'width', d, i, j, cp);
    height = renderer.getParamValue(cp.axisrotated ? 'width' : 'height', d, i, j, cp);
    return {
        x: x,
        y: y,
        width: width,
        height: height
    };
};

svg_plot_renderer.ganttHelpers.getLevelMarkerParams = function(d, i, j, renderer, cp) {
    var rp = renderer.commonRendererProp().params,
        chartObj = rp.chartObj,
        yaxiscolumnorder = rp.yaxiscolumnorder,
        axisrotated = rp.axisrotated,
        dataset = rp.dataset,
        maxRange = cp && cp.yscale.range()[1];

    var percentComplete = d[cp.levelMarker.dataindex],
        yVal = dataset.getY(d, yaxiscolumnorder),
        startVal = yVal[0],
        endVal = yVal[1],
        diff = endVal - startVal,
        marginIncrease = (diff * percentComplete) / 100,
        finalVal = cp.yscale(startVal + marginIncrease),
        xPadding = axisrotated ? 0 : cp.levelMarkerPadding,
        yPadding = axisrotated ? cp.levelMarkerPadding : 0;

    var x, y, wid, hei;
    x = chartObj.dataObject.getTransRectValue('x', d, i, j, cp) + xPadding;
    var basevalue = cp.hasNegValue == null ? cp.yscale(dataset.getY(d, cp.yaxiscolumnorder, 0)) : cp.yscale(cp.hasNegValue);
    if (cp.axisrotated) {
        y = chartObj.dataObject.getTransRectValue('y', d, i, j, cp) + yPadding;
        wid = cp.stacked
            ? mathAbs(cp.yscale(d.y + d.y0) - cp.yscale(d.y0)) + cp.addRangeBand(cp.yscale)
            : (mathAbs(maxRange - basevalue - (maxRange - finalVal)) || 0) + cp.addRangeBand(cp.yscale);
        hei = cp.levelMarkerWidth || cp.barWidth;
    } else {
        y = finalVal;
        wid = cp.levelMarkerWidth || cp.barWidth;
        hei = cp.stacked
            ? mathAbs(basevalue - cp.yscale(d.y + d.y0) - (basevalue - cp.yscale(d.y0))) + cp.addRangeBand(cp.yscale)
            : mathAbs(basevalue - finalVal) + cp.addRangeBand(cp.yscale);
    }
    return {
        x: x,
        y: y,
        width: wid,
        height: hei
    };
}

export var _ganttDependency = plot.render.ganttHelpers;
