// @zohocharts/zc/src/gantt/gantt-connector-encode.js
import { getBarMarkerSize } from '../bar/bar-marker-helpers';
import { defined, pick } from '../utils/zcutils';

export let gantt_connector_position = function() {
    let position_helpers = {};

    const KEY = 'dependency';

    if (this.builderC.commonPlotOptions[KEY] && this.builderC.commonPlotOptions[KEY].enabled) {

        let builderC = this.builderC,
            _this = this,
            dependencyOptions = builderC.commonPlotOptions[KEY], markerOptions;
        if(dependencyOptions.marker) markerOptions = dependencyOptions.marker;

        let dependencyConfig = this.builderC.commonPlotOptions.dependency,
            renderer = _this.oldRenderer;

        this.helpers.position[KEY] = {
            start: (d, encode, pos) => {},
            end: (d, encode, pos) => {

                if (encode == 'size') {
                    return markerOptions.size * 10 || 10;
                }
                else if(encode == 'details') {
                    var dependencyDetails = getDependencyDetails(_this, d, dependencyConfig, renderer, 0);
                    (dependencyDetails = this.helpers.positionComponent[KEY].coordinates(d, this.builderC.datasetIndex, dependencyDetails));
                    return dependencyDetails;
                }
                return {};
            }
        }
    }

    return position_helpers;

};

export let gantt_connector_styles = function() {
    const KEY = 'dependency';

    let _this = this,
        rp = _this.oldRenderer.commonRendererProp().params,
        dataset = rp.dataset,
        cp = _this.oldRenderer.commonOption,
        yaxiscolumnorder = cp.yaxiscolumnorder;

    this.helpers.positionComponent[KEY] = {
        coordinates: (d, i, defaultCoordinates = dependencyDetails) => {
            return defaultCoordinates
        },
        pathString: (d, i, defaultPathString = defaultPathString, type) => {
            return defaultPathString
        }
    }

    if(this.builderC.commonPlotOptions[KEY]?.enabled) {

        let dependencyConfig = this.builderC.commonPlotOptions.dependency,
            linkConf = dependencyConfig.link,
            symConf = dependencyConfig.marker;

        this.helpers.styles[KEY] = {};

        this.helpers.styles[KEY] = {
            fill: 'none',
            stroke: function(d, i) {
                var startTime = dataset.getY(d.source, yaxiscolumnorder)[0];
                var endTime = dataset.getY(d.target, yaxiscolumnorder)[1];
                var error = startTime < endTime;
                return error ? linkConf.errorStyle : linkConf.strokeColor || 'black';
                // return linkConf.strokeColor || 'black';
            },
            strokeWidth: linkConf.strokeWidth,
            strokeDasharray: null,
            strokeLinecap: null,
            display: (d, i) => {
                return '';
            },
            marker: {
                fill: symConf.color,
                stroke: symConf.color,
                strokeWidth: pick(symConf.strokeWidth, 1),
                strokeDasharray: symConf.dashStyle,
                strokeLinecap: symConf.lineCap,
                display: (d, i) => {
                    return '';
                }
            }
        }
    }
};


function getDependencyDetails(_this, d, dependencyConfig, renderer, i) {
    return {
        start: getPos(_this, d.source, renderer, i),
        end: getPos(_this, d.target, renderer, i),
        type: dependencyConfig.link.type.toUpperCase() || 'ES'
    };
}


function getPos(_this, d, renderer, i) {
    var width,
        height,
        x,
        y,
        cp = renderer.commonOption;
        x = _this.chart.dataObject.getTransRectValue('x', d, i, null, cp);
        y = _this.chart.dataObject.getTransRectValue('y', d, i, null, cp);
        width = _this.chart.dataObject.getTransRectValue('width', d, i, null, cp);
        height = _this.chart.dataObject.getTransRectValue('height', d, i, null, cp);
    return {
        x: x,
        y: y,
        width: width,
        height: height
    };
}
