// @zohocharts/zc/src/gantt/ganttOptions.js
import { defaultTheme as options } from '../core/defaulttheme';

options.gantt = function() {
    return {
        outerPadding: 0.04,
        padding: 0,
        interPadding: 0.05,
        maxBandWidth: 200,
        levelMarker: {
            enabled: false,
            dataindex: 2,
            fillOpacity: 0.3,
            bandWidth: 4,
            animation: {
                type: "verticalAll",
                easingType: 'linear',
                enabled: false,
                duration: 300,
                delay: 0,
                exitduration: 0
            }
        },
        hoveroptions: {
            enabled: false,
            start: {
                innerStrokeOpacity: 1,
                outerFillColor: 'white',
                outerStrokeWidth: 2
            },
            end: {
                innerStrokeOpacity: 1,
                outerFillColor: 'white',
                outerStrokeWidth: 2
            }
        },
        marker: {
            enabled: false,
            start: {
                size: 1,
                innerStrokeWidth: 0,
                outerStrokeWidth: 2
            },
            end: {
                size: 1,
                innerStrokeWidth: 0,
                outerStrokeWidth: 2
            }
        },
        border: {
            show: false,
            size: null,
            style: 'solid',
            radius: 0,
            color: null
        },
        stacked: {
            enabled: false,
            showAs: 'values' //values,percent
        },
        animation: {
            type: 'verticalAll'
        },
        datalabels: {
            showAs: 'y' //x,y,percent
        },
        gradients: {
            // linear || none
            options: {
                linear: {
                    y2: 50
                }
            }
        },
        dependency: {
            enabled: false,
            targetIndex: 0,
            link: {
                type: 'ES',
                strokeWidth: 2,
                strokeColor: '#111111',
                errorStyle: 'red',
                padding: 20,
                mode: 'linear'
            },
            marker: {
                enabled: true,
                padding: 10,
                size: 1.5,
                innerFillColor: 'transparent',
                innerStrokeWidth: 0,
                innerStrokeOpacity: 0,
                outerStrokeOpacity: 0,
                outerStrokeColor: 'red',
                outerFillColor: 'blue'
            }
        }
    };
};

export var _ganttOptions = options.gantt;
