// @zohocharts/zc/src/gantt/markerOptions.js
export function getMarkerConfig(config, animationOptions, symbol) {
    return {
        enabled: config.enabled,
        dataindex: config.dataindex,
        columnindex: config.columnindex,
        size: config.size,
        symbol: config.symbol || symbol || 'line',
        innerFillColor: 'transparent',
        innerStrokeColor: 'transparent',
        innerStrokeWidth: 0,
        innerStrokeOpacity: 0,
        outerFillColor: config.color,
        outerStrokeOpacity: config.opacity,
        outerStrokeWidth: config.strokeWidth,
        outerStrokeColor: config.color,
        animation: animationOptions,
        hoveroptions: {
            size: config.hoveroptions.size || config.size,
            symbol: config.symbol || symbol || 'line',
            innerFillColor: config.hoveroptions.innerFillColor || 'transparent',
            innerStrokeColor: 'transparent',
            innerStrokeWidth: 0,
            innerStrokeOpacity: 0,
            outerFillColor: config.hoveroptions.color || config.color,
            outerStrokeOpacity: config.hoveroptions.opacity || config.opacity,
            outerStrokeWidth: config.hoveroptions.strokeWidth || config.strokeWidth,
            outerStrokeColor: config.hoveroptions.color || config.color
        }
    };
}
