// @zohocharts/zc/src/axis-parser/axis-parser.js
import { getAxisObject } from '../axis/axisutils';
import { defined } from '../utils/zcutils';
import { AxisRegistry } from '../axis/axis-registry';

export let axis_parser_helpers = function(chart, config) {
    let processor = chart.dataObject,
        dataset = chart.dataset,
        rotated = chart.axes.rotated,
        hierarchical = processor.hierarchical,
        pivot_helpers;
    if (hierarchical) {
        pivot_helpers = pivot_axis_parser_helpers(chart);
    }

    let helpers = {
        order: (type, order) => {
            if (hierarchical) {
                return pivot_helpers.order(type, order);
            }
            return order;
        },
        orient: (type, order) => {
            if (config.orient) {
                return config.orient;
            }
            if (hierarchical) {
                return pivot_helpers.orient(type, order);
            }
            if (type == 'x') {
                if (order == 0) {
                    return rotated ? 'left' : 'bottom';
                } else {
                    return rotated ? 'right' : 'top';
                }
            } else {
                if (order == 0) {
                    return rotated ? 'bottom' : 'left';
                } else {
                    return rotated ? 'top' : 'right';
                }
            }
        },
        datatype: (type, order) => dataset.getDataType(type, order),
        column: (type, order) => {
            let columnIdx = dataset.getColumnIdx(type, order),
                columns = chart.dataSplicer.getColumns();
            return columns[columnIdx];
        },
        axisInfo: (type, order) => getAxisObject(chart, type, order),
        rotated
    };
    return helpers;
};
var getPriorityValue1 = function(constant, max) {
    if (defined(constant) && defined(constant)) {
        return constant > max ? max : constant;
    } else {
        return constant || max;
    }
};
let pivot_axis_parser_helpers = function(chart) {
    let processor = chart.dataObject,
        hierarchical = processor.hierarchical,
        isCategoryHierarchy = hierarchical && hierarchical.categoryTree,
        pLayout = hierarchical.pivotLayout,
        rowList = pLayout.getRootNodeList('row'),
        columnList = pLayout.getRootNodeList('column'),
        leavesCount = rowList.getLeavesCount(),
        clustersCount = rowList.getClustersCount(),
        splitLeavesColumn = columnList.splitColumnLeaves,
        hasRows = isCategoryHierarchy && hierarchical.hasRows(),
        rotated = chart.axes.rotated;

    let helpers = {
        order: (type, order) => {
            return type == 'y' && isCategoryHierarchy
                ? rowList.hasMultiCluster() && order < leavesCount
                    ? order % clustersCount
                    : 0
                : order;
        },
        orient: (type, order) => {
            if (type == 'x') {
                return rotated ? 'left' : 'bottom';
            } else {
                if (order == 0 || hasRows) {
                    return rotated ? 'bottom' : 'left';
                } else {
                    return rotated ? 'top' : 'right';
                }
            }
        }
    };

    return helpers;
};

AxisRegistry.add('axisH', axis_parser_helpers);
