// @zohocharts/zc/src/axis-parser/axislabel-parser.js
import { defined, mergeJSON } from '../utils/zcutils';
import { dimension_based_intelligence } from '../utils/intelligence-utils';
import { isBranchAxis, isFacet, showAxisLabelForTree } from '../axis/hierarchy-axis-utils';
import { getMargin } from '../axis/axisutils';
import { AxisRegistry } from '../axis/axis-registry';

export function axislabelParser(tickH) {
    let { type, order, chart, config, plotRect, datatype, column, orient } = this,
        { dataObject, dataset, systemConf } = chart,
        axislabeldata = config.label;

    let axislabel_parser_helpers = () => {
        let axislabel_helpers = {
            hideAxislabel: () => {
                let intelligenceDim = chart.canvasarea[orient == 'bottom' || orient == 'top' ? 'height' : 'width'];
                // return;
                return (
                    axislabeldata.show === false ||
                    !defined(axislabeldata.text) ||
                    axislabeldata.text === '' ||
                    !dimension_based_intelligence(chart, `${type}axislabel`, intelligenceDim) ||
                    (tickH.isHierarchy() && !showAxisLabelForTree(type, order, chart)) ||
                    (isFacet(chart) && // Pivot web chart: if no branches are there, hide the axis label
                        dataObject.isPolarAxisCategory &&
                        type == 'y' &&
                        !dataObject.hierarchical.pivotLayout.getRootNodeList('row').branches.length)
                );
            },
            axislabelData: () => {
                let hierarchical = dataObject.hierarchical,
                    isCategoryHierarchy = hierarchical && hierarchical.categoryTree;
                if (type == 'y' && isCategoryHierarchy) {
                    var isBranchAxisType = isBranchAxis(type, order, chart);
                    if (isBranchAxisType) {
                        var branchLabel = {
                            label: {
                                text: column.columnname
                            }
                        };
                        return mergeJSON(config, branchLabel, true);
                    }
                } else {
                    return axislabeldata;
                }
            },
            margin: () => getMargin(config.label),
            branchLabelData: () => {
                let hierarchical = dataObject.hierarchical;
                var branchLabelData;
                if (type == 'y' && hierarchical && hierarchical.categoryTree) {
                    var isBranchAxisType = isBranchAxis(type, order, chart);
                    if (isBranchAxisType) {
                        var branchLabel = {
                            label: {
                                text: column.columnname
                            }
                        };
                        branchLabelData = mergeJSON(config, branchLabel, true);
                    }
                }
                return branchLabelData || config;
            }
        };
        return axislabel_helpers;
    };
    return axislabel_parser_helpers();
}

AxisRegistry.add('axisTitleH', axislabelParser);
