// @zohocharts/zc/src/axis-parser/tick-parser.js
import { NUMERIC_DATATYPE, ORDINAL_DATATYPE, ORDINAL_NON_LINEAR_DATATYPE, TIME_DATATYPE } from '../core/datatype-constants';
import { _linearDomain } from './tick-parser-helpers/tick-domain-helpers/tick-linear-helpers';
import { _ordinalDomain } from './tick-parser-helpers/tick-domain-helpers/tick-ordinal-helpers';
import { _numericSize, _plotNPivotGridSize, _gridSize } from './tick-parser-helpers/grid-size-helpers';
import { getPropVal, getValByPriority, isArray, pick, utils, defined } from '../utils/zcutils';
import {
    getAxisObject,
    isRotatedChart,
    isTickInnerOrient,
    merged_axis_column_intersect,
    hasMergedAxis,
    getMargin,
    getTickPadding
} from '../axis/axisutils';
import { getFormattedValue, getFormattedValueByPercent, getAutoTimeSubFunction } from '../formatter/formatter';
import { LOG } from '../core/scale-type-constant';
import { $ZCG } from '../core/zcglobal';
import { hasBranchAxis, isBranchAxis } from '../axis/hierarchy-axis-utils';
import { getMaxTickSize } from '../axis/tick-helpers';
import { _autoRotation } from './tick-parser-helpers/tick-rotation-helpers';
import { _tickPosition } from './tick-parser-helpers/tick-position-helpers';
import { getScaleForTicks } from './tick-parser-helpers/tick-scale-helpers';
import { formatTickWrapByMode } from '../axis/tickWrap';
// import { _logDomain } from './tick-parser-helpers/tick-domain-helpers/tick-log-helpers';
import { getD3Scale } from '../axis/scale';
import { DataProcessor as P } from '../core/data-processor';
import { pivot_checkers } from './tick-parser-helpers/tick-hierarchy-helpers';
import { ticklabelGetter } from './tick-parser-helpers/ticklabel-getter';
import { AxisRegistry } from '../axis/axis-registry';
import { scale } from '../core/main-zc-init';

export function tickParser(helpers) {
    let { type, order, chart, config, plotRect, datatype, column } = helpers,
        { dataObject, dataset, systemConf } = chart;

    let originalDataType = column.datatype,
        hierachical = dataObject.hierarchical,
        scaleType = column.scaleType,
        isLogScale = scaleType === LOG,
        longestTick,
        side = isRotatedChart(type, chart.axes.rotated) ? 'height' : 'width',
        maxTickSize = getMaxTickSize(config.ticklabel, plotRect),
        axisInfo = getAxisObject(chart, type, order);

    let _h = tick_parser_helpers({ ...helpers, side });

    return _h;
}

let tick_parser_helpers = helpers => {
    let { chart, type, config, plotRect, datatype, order, maxSize } = helpers,
        { dataObject, dataset, systemConf } = chart,
        processor = chart.dataObject,
        hierarchical = processor.hierarchical,
        oldTickInfo = _existingPrerenderComp(helpers),
        firstChartName = chart.chartTypes[0].name,
        commonPlotOptions = chart.systemConf.chart.plot.plotoptions[firstChartName],
        pivot_helpers,
        _pivot_checkers = pivot_checkers(helpers),
        _tick_getter = ticklabelGetter(helpers, _pivot_checkers);
    if (hierarchical) {
        pivot_helpers = pivot_tick_parser_helpers(helpers);
    }
    let tick_helpers = {
        domain: () => _domain(helpers),
        scale: extent => _scale(helpers, null, extent),
        simpleScale: () => _scale(helpers, true),
        size: () => _size(helpers),
        orient: () => _orient(helpers),
        filter: () => helpers.chart.renderVisiblePointsOnly,
        useTickAsIndices: hierarchical ? pivot_helpers.useTickAsIndices : datatype == ORDINAL_DATATYPE && processor.isWaterfallEnabled,
        styles: () => _styles(helpers),
        rotation: () => _autoRotation(helpers),
        format: () => _format(helpers),
        formatter: () => _formatter(helpers),
        formatAccessor: _formatAccessor,
        side: () => _side(chart, type),
        groupby: () => _hasGroupby(helpers),
        maxSize: () => (oldTickInfo ? oldTickInfo.maxTickSize : getMaxTickSize(config.ticklabel, plotRect)),
        autoScrollX: () =>
            (datatype == ORDINAL_DATATYPE || datatype == ORDINAL_NON_LINEAR_DATATYPE) &&
            (systemConf.chart.width === 'auto' || systemConf.chart.height === 'auto'),
        autoScrollY: () => (systemConf.chart.width === 'auto' || systemConf.chart.height === 'auto') && hierarchical && type == 'y',
        chartPadding: () => (dataObject.isAxisCategory ? commonPlotOptions.padding || 0.2 : 0),
        chartOuterPadding: () => (dataObject.isAxisCategory ? commonPlotOptions.outerPadding || 0 : 0),
        logTickH: () => scale.log.domain(helpers),
        scrollThickness: () => getPropVal(chart.systemConf, 'chart.scroll.categoryThickness'),
        chartRect: () => chart.chartarea,
        plotSize: () => _gridSize(helpers, true),
        plotSize_no_zoom: () => _gridSize(helpers, false),
        bzsEnabled: () => _BZS(helpers),
        filteredTickIndices: ticks => processor.isWaterfallEnabled && _existingScale(helpers) && isArray(ticks),
        subFormatType: _subFormatType,
        margin: () => getMargin(config.ticklabel),
        padding: margin => getTickPadding(helpers.orient, margin, isTickInnerOrient(config, type, chart)),
        getActualTicklabel: _tick_getter.getActualTicklabel,
        getTicksforAxisComp: _tick_getter.getTicksforAxisComp,
        isHierarchy: _pivot_checkers.isHierarchy,
        isHierarchialLevelType: _pivot_checkers.isHierarchialLevelType
    };
    return { ...tick_helpers, ...pivot_helpers };
};
function _side(chart, type) {
    return isRotatedChart(type, chart.axes.rotated) ? 'height' : 'width';
}

function pivot_tick_parser_helpers(helpers) {
    let { type, order, chart } = helpers,
        { dataObject, systemConf } = chart;

    let processor = chart.dataObject,
        hierarchical = processor.hierarchical,
        isCategoryHierarchy = hierarchical && hierarchical.categoryTree,
        pLayout = hierarchical.pivotLayout,
        rowList = pLayout.getRootNodeList('row'),
        columnList = pLayout.getRootNodeList('column'),
        splitLeavesColumn = columnList.splitColumnLeaves;

    let pivot_tick_helpers = {
        useTickAsIndices: hierarchical.categoryTree && type == 'x' && !splitLeavesColumn && order == 0,
        autoScrollY: () => {
            let domainLength = rowList.count ? rowList.getDomainForBranch(0).length : 0;
            return (
                (systemConf.chart.width === 'auto' || systemConf.chart.height === 'auto') &&
                type == 'y' &&
                rowList.count &&
                order == domainLength - 1
            );
        },
        pivotLeafSize: () => {
            return { width: 0, height: rowList.leafSize };
        },
        pivotLeafCount: () => rowList.getLeavesCount(),
        isBranchType: () => (isCategoryHierarchy ? isBranchAxis(type, order, chart) : type == 'x' && order > 0),
        isLeafType: () => {
            let isBranchAxisType = isCategoryHierarchy ? isBranchAxis(type, order, chart) : type == 'x' && order > 0;
            return (
                (isCategoryHierarchy && type == 'x' && !isBranchAxisType && hasBranchAxis(type, chart)) ||
                (!isCategoryHierarchy && type == 'x' && chart.axes.x.count > 1)
            );
        }
    };
    return pivot_tick_helpers;
}

export function _BZS(helpers) {
    let { chart } = helpers,
        behaviours = getPropVal(chart, 'behaviouralEvents.eventManager.behaviours') || {},
        _brush_zoom_scroll_enabled = behaviours.PLOT_ZOOM || behaviours.PLOT_BRUSH || behaviours.CHART_SCROLL;
    return _brush_zoom_scroll_enabled;
}
function _scale(helpers, simpleD3Mode, extent) {
    let { chart, type, order, column } = helpers,
        axisInfo = getAxisObject(chart, type, order),
        behaviours = getPropVal(chart, 'behaviouralEvents.eventManager.behaviours') || {},
        scale;
    if (behaviours.PLOT_ZOOM || behaviours.PLOT_BRUSH || behaviours.CHART_SCROLL) {
        scale = axisInfo.scale;
    }
    if (!scale) {
        if (simpleD3Mode) {
            let scaleType = column.scaleType || 'linear';
            scale = getD3Scale(scaleType);
            scale.isD3 = true;
        } else {
            scale = getScaleForTicks(helpers, extent);
        }
    }
    return scale;
}
export function _existingPrerenderComp(helpers) {
    return helpers.axisInfo.prerenderComp && helpers.axisInfo.prerenderComp.axisTicks.tickInfo;
}
export function _existingScale(helpers) {
    let { chart, type, order } = helpers,
        axisInfo = getAxisObject(chart, type, order),
        scale = axisInfo.scale;
    return scale;
}

function _orient() {}

function _domain(helpers) {
    if (helpers.datatype == NUMERIC_DATATYPE || helpers.datatype == TIME_DATATYPE) {
        return _linearDomain(helpers);
    } else {
        return _ordinalDomain(helpers);
    }
}

function _size(helpers) {
    if (helpers.datatype == NUMERIC_DATATYPE) {
        return _numericSize(helpers);
    } else if (helpers.datatype == TIME_DATATYPE) {
        return _plotNPivotGridSize(helpers); //TODO code cleanup
    } else if (helpers.datatype == ORDINAL_DATATYPE || helpers.datatype == ORDINAL_NON_LINEAR_DATATYPE) {
        return _plotNPivotGridSize(helpers); //TODO code cleanup
    }
}

function _styles(helpers) {
    let tickConfig = helpers.config.ticklabel,
        systemConf = helpers.chart.systemConf;
    return {
        fontSize: parseFloat(utils.getVal(getValByPriority([tickConfig, systemConf.chart, systemConf.canvas, $ZCG], 'fontSize'))),
        fontFamily: utils.getVal(getValByPriority([tickConfig, systemConf.chart, systemConf.canvas, $ZCG], 'fontFamily'))
    };
}

function _hasGroupby(helpers) {
    let formatConfig = _format(helpers),
        hasTimeGroupBy =
            (isArray(formatConfig.groupby) && formatConfig.groupby.length > 0) ||
            (formatConfig.groupby === 'AUTO' && formatConfig.subfunction !== 'YEAR');

    if (hasTimeGroupBy && formatConfig.groupby === 'AUTO') {
        //#ZC1015
        formatConfig.AUTO_GROUP_ORDER = P.HierarchicalTimeDataAdapter.getAutoGroupOrder(formatConfig.subfunction).reverse();
    }
    return hasTimeGroupBy;
}

function _format(helpers) {
    let { type, order, chart, column } = helpers,
        isMergedAxis = hasMergedAxis(helpers),
        formatConfig = pick(isMergedAxis ? merged_axis_column_intersect(order, chart) : column[column.datatype], {});

    return formatConfig;
}

function _formatter(helpers) {
    let { type, order, chart, datatype } = helpers;
    return type === 'y' && chart.dataObject.isPercentChart ? getFormattedValueByPercent : getFormattedValue;
}

function _formatAccessor(helpers, tickH) {
    let { type, order, chart, datatype, column } = helpers,
        { formatter, maxSize, styles, format, groupby, rotationConfig } = tickH,
        needTickConfigWrap = rotationConfig.tickAlignmentMode == 'wrapOnSpace' || rotationConfig.tickAlignmentMode == 'wrapAll',
        specifierIndex = groupby ? order : null;
    return function(value) {
        var formattedValue = formatter(format, value, column.datatype, chart, 'axis', null, null, specifierIndex);
        formattedValue = needTickConfigWrap
            ? formatTickWrapByMode(formattedValue, rotationConfig.tickAlignmentMode, maxSize.horizontal, maxSize.vertical, styles.fontSize)
            : formattedValue;
        return formattedValue;
    };
}

function _subFormatType(formatConfig, timeData) {
    let _subFormatType = formatConfig.subfunction;
    if (!defined(_subFormatType) || _subFormatType === 'AUTO') {
        // set sub function
        var info = getAutoTimeSubFunction(timeData);
        _subFormatType = formatConfig.subfunction = info.subFunction;
    }
    return _subFormatType;
}

AxisRegistry.add('tickH', tickParser);
