// @zohocharts/zc/src/behaviours/legend-brush/color-band/legend-color-band-brush-arguments.js
//
import { Registry } from '../../../core/zc-registry';
import { TRANSPARENT } from '../../../core/zc-constants';

/**
 *
 * @param chart
 * @return {BrushD3Options}
 */
export function _getLegendColorBandBrushArguments(chart) {
    let legendInstance = chart.legends.get('clr-0'),
        { component, element } = legendInstance,
        { config } = component.spec,
        { type, dimension } = component.brushOptions,
        //
        extent = [[0, 0], dimension],
        BrushD3Options = Registry.getComponent('BrushD3Options');

    return new BrushD3Options({
        type,
        extent,
        component,
        behaviour: 'resizeToMouse',
        element: getTrackerElement(element),
        styles: {
            selection: {
                fillColor: TRANSPARENT,
                strokeColor: TRANSPARENT,
                fillOpacity: 1
            },
            track: config.filter.track
        }
    });
}

/**
 * Returns tracker element
 * @param  {D3Selectiion} container
 * @return {D3Selectiion}
 */
function getTrackerElement(container) {
    return container.selectAll('g.zc-band-brush');
}

Registry.setComponent('_getLegendColorBandBrushArguments', _getLegendColorBandBrushArguments);
