// @zohocharts/zc/src/behaviours/legend-brush/color-band/legend-color-band-brush-handler-mappings.js
//
import { LegendColorBandBrushListener } from '../../listeners/legend/legend-color-band-brush-listener';
import { LegendBrushDataFinder } from '../../data-selection/legend/legend-brush-data-finder';
import { BAND_LEGEND_TYPE } from '../../../legend/legend-type-constants';
import { Registry } from '../../../core/zc-registry';
import { d3_set } from '../../../core/d3-constants';
import { validateNullVal } from '../../../utils/zcutils';
import { getHandles } from './handles';
import { brushTooltip } from '../size-slider/brush-tooltip';
import { handleShadow } from '../../visual-effects/brush-handles/handle-shadow';
import { isSupported } from '../../../legend/events/legend-event-helpers';

import { StyleHandles } from '../../visual-effects/legend/color-band/style-handles';
import { ApplyBrushSelectionOnPlot } from '../../visual-effects/legend/size-slider/update-plot-by-selection';
import { showTooltip } from '../../visual-effects/legend/size-slider/show-tooltip';
import { hideTooltip } from '../../visual-effects/legend/size-slider/hide-tooltip';
import { HideHighlight } from '../../visual-effects/legend/size-slider/hide-highlight';
import { PreventDefault } from '../../visual-effects/legend/size-slider/slider-prevent-default';
import { SnapRange } from '../../visual-effects/legend/size-slider/snap-range';
import { ForceHideTooltip } from '../../visual-effects/legend/size-slider/force-hide-tooltip';
import { ClearCache } from '../../../legend/events/visual-effects/clear-cache';
import { StyleTrack } from '../../visual-effects/legend/color-band/style-track';

let _legendColorBandBrushHandlers = {
    LISTENER: {
        start: LegendColorBandBrushListener,
        brush: LegendColorBandBrushListener,
        end: LegendColorBandBrushListener
    },
    DATA_FINDER: {
        start: LegendBrushDataFinder,
        brush: LegendBrushDataFinder,
        end: LegendBrushDataFinder
    },
    BEHAVIOUR: {
        start: [PreventDefault],
        brush: [ForceHideTooltip, StyleHandles, ApplyBrushSelectionOnPlot, showTooltip, HideHighlight, StyleTrack],
        end: [ClearCache, hideTooltip, SnapRange]
    },
    LISTENER_TYPES: d3_set(['start', 'brush', 'end']),
    SUPPORTED: isSupported(BAND_LEGEND_TYPE),
    // UPDATE: function() {},
    INIT: function(behaviour) {
        let behaviouralEvents = this,
            chart = behaviouralEvents.chart,
            behaviourInstance = behaviour.instance,
            legendComponent = behaviourInstance.options.component,
            { config } = legendComponent.spec,
            { initialExtent } = legendComponent.brushOptions,
            handles = getHandles(behaviour),
            eventEnabled = isEventEnabled(config);

        behaviour.initialized = false;
        let handleElement = behaviourInstance.addCustomHandle(handles);
        behaviourInstance.setSelection(initialExtent);
        behaviourInstance.tooltip = brushTooltip(behaviour);

        // Shadow
        let filterConfig = config.filter.slider;
        handleShadow(filterConfig, handles, handleElement, chart);

        // Visibility for custom handles
        behaviourInstance.customHandles.style('visibility', eventEnabled ? null : 'hidden');

        if (!eventEnabled) {
            behaviourInstance.customHandles.attr('pointer-events', 'none');
            behaviourInstance.destroy();
        }

        behaviour.initialized = true;
    }
};

function isEventEnabled(config) {
    let { events: eventsConfig, filter } = config,
        eventDisabled = eventsConfig && validateNullVal(eventsConfig.enabled, true) == false;

    return !eventDisabled && filter.enabled;
}

Registry.setComponent('_legendColorBandBrushHandlers', _legendColorBandBrushHandlers);

export { _legendColorBandBrushHandlers };
