// @zohocharts/zc/src/legend/color-legend/color-legend.js
//
import { ORDINAL_DATATYPE } from '../../core/datatype-constants';
import { BaseLegend } from '../base-legend';
import { legend, legend_helpers, scale } from '../../core/main-zc-init';
import { d3_range } from '../../core/d3-constants';
import { getRawRanges, colorscale_helpers } from '../../axis/colorscale-helpers';
import { axis } from '../../core/main-zc-init';
// import { legend_helpers } from '../legend-helper';
import { getMaxToResize } from '../legend-bound';

export class ColorLegend extends BaseLegend {
    create() {
        this.encode = getlegend(this.spec.chartObj, this.spec);
    }

    draw(parent) {
        this.parent = parent;
        let isContinousLegend = legend_helpers.isContinousLegend(this.spec.chartObj);
        if (isContinousLegend) {
            super.draw(parent);
            return;
        }

        this.encode(parent.datum(this.spec.config));
    }

    resize(width, height) {
        if (!width && !height) {
            return;
        }

        let bound = this.spec.bound,
            { w, h } = getMaxToResize(width, height, bound);
        bound.maxWidth = w;
        bound.maxHeight = h;

        this.create();
        this.draw(this.parent);
    }
}

export function createColorAxis(chartObj) {
    let isQuantileColorScale = colorscale_helpers.isQuantileColorScale(chartObj);
    if (isQuantileColorScale) {
        let userData = chartObj.userdata,
            config = chartObj.systemConf.legend,
            dataType = chartObj.dataset.getDataType('clr');

        if (dataType === ORDINAL_DATATYPE) {
            userData.legend = userData.legend || {};
            userData.legend.colorBand = userData.legend.colorBand || {};
            if (!userData.legend.colorBand.ranges) {
                config.colorBand.ranges = userData.legend.colorBand.ranges = colorscale_helpers.getCategories(chartObj);
            }
        }

        var userRanges = config.colorBand.ranges;
        if (userRanges && userRanges.length) {
            userData.legend.colorBand.ranges = config.colorBand.ranges = getRawRanges(config, chartObj, true); //for  [10,20,30,40] ranges where the ranges are not array, it is mandatory to convert it to array and save it in userdata and systemconf for performing legend highlight, filter and redraw after legend filter
        }

        config.ranges = config.colorBand.ranges;

        // Legend type switch with filter.range
        // if (!legend_helpers.isContinousLegend(chartObj)) {
        //     let userRanges = config.colorBand.ranges;
        //
        //     if (config.filter.range) {
        //         let activeDomain = getActiveDomain(config, chartObj),
        //             activeIndices = activeDomain.map(_ => arrayFindIndex(userRanges, range => jsonStringify(_) === jsonStringify(range)));
        //
        //         userRanges.forEach((item, i) => (item.disabled = !isWithinRange(i, activeIndices)));
        //         config.filter.range = null;
        //     }
        // }

        var colorScale = scale.getScale(chartObj, chartObj.dataObject, null, null, null, chartObj.dataset, 'clr'),
            colorScale_domain = colorScale.domain(),
            indexScale = colorScale.copy();
        indexScale.range(
            colorscale_helpers.getProcessedColorRange(d3_range(colorScale_domain.length), colorScale_domain, config, chartObj)
        );

        var axisObj = chartObj.axes.clr || {};
        axisObj.scaleType = 'quantile';
        axisObj.dataType = dataType;
        axisObj.scale = colorScale;
        axisObj.indexScale = indexScale;
        axisObj.schema = { data: { meta: { ...chartObj.dataset.getColumn('clr'), columnindex: chartObj.dataset.getColumnIdx('clr') } } };
        axisObj.component = axisObj.component || axis.get(); //#ZC1310
        chartObj.axes.clr = axisObj;
    } else {
        chartObj.axes.clr = {};
    }
}
// Registry.setComponent('createColorAxis', createColorAxis);

function getlegend(chartObj, spec) {
    let legComp = (chartObj.legend.component = legend.discrete.get(spec));

    var legend_comp = legComp
        .chartarea(chartObj.chartarea)
        .systemconf(chartObj.systemConf)
        .color(chartObj.seriesColor)
        .chartObj(chartObj);
    return legend_comp;
}
