// @zohocharts/zc/src/components/race/handlers/continuousRace-handler.js
import { d3_scaleLinear } from '../../../core/d3-constants';
import { defined, isNull, isUndefined } from '../../../utils/zcutils';

export function continuousRaceHandler(datatype, _rp) {
    //without scroll not handled, toa chieve it we need to set scroll height ot 0 and slider size to 0.
    // var ordinalScaleToFindPos;
    // if (datatype === 'ordinal') {
    //     ordinalScaleToFindPos = getOrdinalScaleToFindPos.call(this);
    // }


    var continuousRaceInfo = this.continuousRaceInfo,
        chart = this.chart,
        scroll = chart.scroll,
        instance = scroll && scroll.brushInstance,
        range = scroll && scroll.brushScale && scroll.brushScale.range(),
        selection = scroll && scroll.brushInstance && scroll.brushInstance.getSelection();
    if (selection[1] === range[1]) {
        //when end of scroll reached, we need reset the value to intial value.
        continuousRaceInfo.index = continuousRaceInfo.initialValue - 1;
        //resetting when final extent is reached.
    }
    var isOrdinal = datatype === 'ordinal',
        isreplay = chart.racing.replay || instance.getSelection()[1] === chart.scroll.brushScale.range()[1]; //check for auto || manual case
    if (isUndefined(continuousRaceInfo.initialValue)) {
        //gets called on init and on redraw
        //on init, and on every redraw, we need recalculate this, continuousRaceInfo.index will be -1 on redraw.
        continuousRaceInfo.initialValue = getInitialDataCount(chart);
        var maxcount = isOrdinal ? chart.scroll.brushScale.domain().length : chart.racing.continuousRaceInfo.raceScaleX.domain().length;
        if (maxcount <= continuousRaceInfo.initialValue) {
            continuousRaceInfo.initialValue = 1;
        }
    }
    //race index update, each time one index index is increased
    continuousRaceInfo.index++;
    if (!continuousRaceInfo.index) {
        //at first -1 -> 0
        continuousRaceInfo.index = continuousRaceInfo.initialValue;
    } else if (continuousRaceInfo.index < 0) {
        continuousRaceInfo.index = 0;
    }

    //collecting the nextData possible
    var info = getNextDataInfo(chart.seriesdata, continuousRaceInfo.index, chart),
        nextDataMaxPos = info.max,
        nextData;
    //if replayed - need to position faster to initial value, no animation from last point to initial point.
    if (nextDataMaxPos === 0 || isreplay) {
        //when replayed the race here it comes //nextDataMaxPos will be 0 when there is no next data position, since even if continuousRaceInfo.index is 0 we block it before itself.
        continuousRaceInfo.index = 1;
        continuousRaceInfo.index = continuousRaceInfo.index === 0 ? 1 : continuousRaceInfo.index;
        info = getNextDataInfo(chart.seriesdata, continuousRaceInfo.index, chart);
        nextDataMaxPos = info.max;
        chart.racing.syncTransition = chart.container.transition().duration(0);
    }

    
    if (datatype === 'ordinal') {
        nextData = info.maxData;
        chart.racing.scrollView = getRaceScrollView(chart, isOrdinal, nextData);
        var _end = chart.scroll.brushScale(nextData) + chart.scroll.brushScale.bandwidth() / 2;
        if (info.isLastData) {
            _end = chart.scroll.brushScale.range()[1];
        }
        instance.setSelection([instance.getSelection()[0], _end], 0, chart.racing, true);
    } else {
        nextData = info.maxData;
        //scrollViewUpdate
        chart.racing.scrollView = getRaceScrollView(chart, isOrdinal, nextData); //for filtering of data in the plot
        var _end = /*chart.scroll.brushScale(nextData) >= minimumEndVal ?*/ chart.scroll.brushScale(nextData); // : minimumEndVal;

        if (info.isLastData) {
            _end = chart.scroll.brushScale.range()[1];
        }
       instance.setSelection([instance.getSelection()[0], _end], 0, chart.racing, true); //for positioning of scroll slider and xaxis creation(if no minmax)
    }
    updateYaxisDomain(chart);
    //race loop break condition
    if(chart.dataset.getDataType("x") === "ordinal"){
        var breakLoop = (chart.scroll.brushScale.domain().length <= continuousRaceInfo.index + 1) || info.isLastData;
        return () => breakLoop;
    }else{
        var breakLoop = (chart.racing.continuousRaceInfo.raceScaleX.domain().length <= continuousRaceInfo.index + 1) || info.isLastData;
        return () => breakLoop;
    }
    
}
export function updateYaxisDomain(chart, index){
    var continuousRaceInfo = chart.racing.continuousRaceInfo;
    //on redraw we will empty continuousRaceInfo.index, so we depend on index.
    index = defined(index) ? index : continuousRaceInfo.index >= 0 ? continuousRaceInfo.index : null;
    var min = isNull(index) ? null : continuousRaceInfo.yDataMinArr[index],
        max = isNull(index) ? null : continuousRaceInfo.yDataMaxArr[index],
        outerPadding = chart.axes.y.groups[0].scale.outerPadding,
        domain = chart.axes.y.groups[0].scale.domain(),
        systemConf = chart.systemConf,
        isYMin = defined(systemConf.chart.axes.yaxis[0].minRange),
        isYMax = defined(systemConf.chart.axes.yaxis[0].maxRange);

    //if the min/max value given is null, previous domain have to be retained
    domain[0] = !isYMin && min != null ? min - (max - min) * outerPadding : domain[0];
    domain[1] = !isYMax && max != null ? max + (max - min) * outerPadding : domain[1];
    if(min === max){
        domain[0] = chart.axes.y.groups[0].scale.domain()[0];
    }
    chart.racing.interrupt = true;
    chart.updateScale(null, 'y', domain, 0, null, true);
    chart.racing.interrupt = false;
    
}

function getNextDataInfo(data, index, chart) {
    var max,
        maxData,
        isLastData,
        brushdomain = chart.scroll.brushScale.domain();
    if (chart.dataset.getDataType('x') === 'ordinal') {
        maxData = brushdomain[index];
        max = chart.scroll.brushScale(maxData);
        if (index >= brushdomain.length - 1) {
            maxData = brushdomain[brushdomain.length - 1];
            max = chart.scroll.brushScale(maxData);
            isLastData = true;
        }
    } else {
        max = chart.racing.continuousRaceInfo.raceScaleX.invert(index);
        maxData = chart.scroll.brushScale.invert(max);
        if (index >= chart.racing.continuousRaceInfo.raceScaleX.domain().length - 1) {
            maxData = chart.scroll.brushScale.domain()[1];
            max = chart.scroll.brushScale.range()[1];
            isLastData = true;
        }
    }
    
    return { max, maxData, isLastData };
}

export function getMaxSeriesDataCount(data) {
    var maxCount = 0;
    data.map(function(series) {
        maxCount = series.data[0].length > maxCount ? series.data[0].length : maxCount;
    });
    return maxCount;
}

export function getInitialDataCount(chart) {
    var maxCount = 0;
    if(chart.dataset.getDataType("x") == "ordinal"){
        return chart.axes.x.scale.domain().length;
    }
    chart.renderer.map(function(renderer) {
        var seriesdataAfterUpdateLength = renderer.seriesdataAfterUpdate[0].length;
        maxCount = maxCount < seriesdataAfterUpdateLength ? seriesdataAfterUpdateLength : maxCount;
    });
    return maxCount;
}

export function getRaceScrollView(chart, isOrdinal, nextData, isRedraw) {
    var scroll = chart.scroll,
        instance = scroll && scroll.brushInstance,
        race = chart.racing,
        { maxRange } = chart.systemConf.chart.axes.xaxis,
        keyFramedRace = chart.racing.keyFramedRace;

    if(keyFramedRace){
        return null;
    }
    if (isOrdinal) {
        return (basedOnXscale, basedOnScrollScale) => {
            scroll.visiblePointCount = null;
            if (basedOnScrollScale) {
                if (isUndefined(nextData) || isRedraw) {
                    return [instance.getSelection()[0], instance.getSelection()[1]];
                }
                return [instance.getSelection()[0], chart.scroll.brushScale(nextData) + chart.scroll.brushScale.bandwidth() / 2];
            } else if (basedOnXscale) {
                if (isUndefined(nextData) || isRedraw) {
                    return [
                        instance.getSelection()[0],
                        race.isEntireAxisNeededForOrdinal ? instance.getSelection()[1] : chart.axes.x.scale.range()[1]
                    ];
                }
                return [instance.getSelection()[0], chart.axes.x.scale(nextData) + chart.axes.x.scale.bandwidth() / 2];
            }
        };
    } else {
        return (basedOnXscale, basedOnScrollScale) => {
            scroll.visiblePointCount = null;
            if (basedOnScrollScale) {
                //for xaxis
                if (isUndefined(nextData) || isRedraw) {
                    return [instance.getSelection()[0], instance.getSelection()[1]];
                }
                var _end = /*chart.scroll.brushScale(nextData) >= minimumEndVal ?*/ chart.scroll.brushScale(nextData)// : minimumEndVal;
                    return [instance.getSelection()[0], _end];
            } else if (basedOnXscale) {
                //for filtering of data
                var _startD = getNextDataInfo(chart.seriesdata, 0, chart).maxData;
                if (isUndefined(nextData) || isRedraw) {
                    return [chart.axes.x.scale(_startD), !maxRange ? chart.axes.x.scale.range()[1] : instance.getSelection()[1]];
                }
                var _end = /*chart.axes.x.scale(nextData) >= minimumEndVal ?*/ chart.axes.x.scale(nextData)// : minimumEndVal;
                return [chart.axes.x.scale(_startD), _end];
            }
        };
    }
}
