// @zohocharts/zc/src/geo-scatter/geo-scatter-events.js
import { GeoScatterSeries } from './geo-scatter';
import { isDataActive } from '../core/data-processor/is-data-active';
import { isArray } from '../utils/zcutils';
import { LOCATION_MARKERS } from '../geo-core/geo-marker-constant';
import { getGeoEnclosedPoints } from '../geo-core/zc_map_events';
import { pointInRectangle } from '../events/events-utils';

/**
 *
 * @param {array} latLong
 * @param {array} XY
 * @return {*}
 */
GeoScatterSeries.prototype.getPoint = function(latLong, XY) {
    var series = this,
        geoMap = series.geoMap,
        plotArea = geoMap.plotarea,
        dataObject = geoMap.dataObject,
        chartName = geoMap.chartTypes[series.index].name,
        flattenPoints = dataObject.flattenPoints.get(chartName),
        plotOptions = series.plotOptions,
        markerSizeFactor = 10,
        mouseX = XY[0],
        mouseY = XY[1];

    // ZC2358
    if (!isArray(flattenPoints)) {
        return;
    }
    // if the symbol is location type. add half of the size/2
    if (LOCATION_MARKERS.has(plotOptions.marker.symbol)) {
        mouseY += (plotOptions.marker.size * markerSizeFactor) / 2;
    }

    let point = getGeoEnclosedPoints(flattenPoints, 'geoscatter', [mouseX, mouseY], 'QUADTREE', geoMap);

    if (
        point &&
        // #ZC1805
        pointInRectangle(point._x, point._y, {
            x: 0,
            y: 0,
            width: plotArea.width,
            height: plotArea.height
        })
    ) {
        if (!geoMap.hasColorScale || (geoMap.hasColorScale && isDataActive(point, geoMap))) {
            return {
                type: 'point',
                point: point
            };
        }
    }
    return null;
};

export var _geoScatterEvent = GeoScatterSeries.prototype;
