// @zohocharts/zc/src/legend/list-legend/list-legend.js
//
import { Registry } from '../../core/zc-registry';
import { BaseView } from '../../view/base-view';
import { getVisibleItemRenderCount } from '../size-legend/list/expandable-helper';
import { ListLegendParser } from './list-legend-parser';

let EMPTY_RETURN_FUNCTION = () => ({});

/*
spec = {
    data: ['North', 'South', 'East', 'West'],
    bound: {
        maxWidth: 150,
        maxHeight: 150
    },
    config: { // Size list config
        list: {
            marker: {
                fillColor: "yellow",
                strokeColor: "red",
                strokeWidth: 2
            },
            label: {
                fontSize: 13,
                padding: 5
            }
        }
    },
    globalFont: {}
}
 */

class ListLegend extends BaseView {
    constructor(spec, viewScaleComponent) {
        super(spec);
        this.viewScaleComponent = viewScaleComponent || EMPTY_RETURN_FUNCTION;
    }

    create() {
        return (this.encode = ListLegendParser(this.spec, this.viewScaleComponent(), this));
    }
}

export { ListLegend };
Registry.setComponent('listLegend', ListLegend);
