// @zohocharts/zc/src/legend/list-legend/list-legend-parser.js
//
import { DIV_ELEMENT } from '../../core/element-constants';
import { HORIZONTAL_ALIGN } from '../../core/align-constants';
import {
    LEGEND_TITLE_MAX_WIDTH_HORIZONTAL,
    LEGEND_TITLE_MAX_WIDTH_VERTICAL,
    LEGSERIES_LEGEND_CLASSNAME,
    SERIES_LEGEND_CLASSNAME,
    LEGSERIES_HEADER_CLASSNAME,
    EXPANDABLE_CLASSNAME
} from '../legend-constants';
import { $ZCG } from '../../core/zcglobal';
import { LegendBackgroundHtmlParser } from '../background/legend-background-html-parser';
import { LegendTitleHtmlParser } from '../title/legend-title-html-parser';
import { ListParser } from '../../view/list/list-parser';
import { ListView } from '../../view/list/list-view';
import { ExpandableParser, expandableState } from '../expandable/expandable-parser';
import { Mark } from '../../mark/mark-utils';
import { mathMin, mathMax, NULL, PX, mathRound } from '../../core/zc-constants';
import { d3_max } from '../../core/d3-constants';
import {
    setPropVal,
    getBoundingRect,
    d3_values,
    defined,
    pick,
    utils,
    getValByPriority,
    perToPx,
    elementIsScrollable
} from '../../utils/zcutils';
import { getLegendAlignment } from '../legend-utils';
import { legend_helpers } from '../../core/main-zc-init';
import { getBoxSize } from '../../view/list/list-helpers';
import { $Browser } from '../../core/browser-detect';
import { ExpandableAdjustFactor } from '../expandable/expandable-adjust';

function ListLegendParser(spec, viewScaleComponent, self) {
    let _helper = helper(spec, viewScaleComponent, self),
        {
            titleSpec,
            backgroundSpec,
            viewSpec,
            expandableSpec,
            keepTitleAligned,
            setDimension,
            keepExpandableAligned,
            initLegendComp
        } = _helper;

    // Wrap
    let groupMark = Mark(DIV_ELEMENT, `.legendgroup`);
    groupMark.onEnd = setDimension;
    groupMark.onStart = initLegendComp;

    // Background
    setPropVal(groupMark, 'properties.styles', LegendBackgroundHtmlParser(backgroundSpec)); // TODO: change by config

    // Title
    let titleMark = LegendTitleHtmlParser.call(spec, titleSpec);
    setPropVal(titleMark, 'properties.styles.display', getDisplay(keepTitleAligned));

    // View
    let listMark = ListParser(viewSpec);
    setPropVal(listMark, 'properties.styles.display', getDisplay(keepTitleAligned, keepExpandableAligned));

    // Expandable
    let expandableMark = ExpandableParser({ ...expandableSpec, target: () => listMark.getElement() });

    groupMark.children = [titleMark, listMark, expandableMark];
    groupMark._helper = _helper;
    return groupMark;
}

function helper(spec, viewScaleComponent, self) {
    let { config, globalFont, format, data, bound, callbackArgs, scales = {}, helpers, chartObj, computedInfo } = spec,
        { title: titleConfig, list: listConfig } = config,
        { expandable: expandableConfig = {} } = listConfig,
        // View scale component
        { scale: viewScale, bound: viewBound, ticks, labelData } = viewScaleComponent,
        // Align
        legendAlign = getLegendAlignment(config),
        isSizeLegend = spec.encode === 'z',
        legData = !isSizeLegend && chartObj ? legend_helpers.getLegendData(chartObj, spec.order) : data, //reason for adding chartObj is for customlegend - DASHBOARD_FLAT_16
        legData_length = !isSizeLegend && chartObj ? legend_helpers.getLegendDataLength(chartObj, legData) : data.length,
        isHorizontal = legendAlign.layout === HORIZONTAL_ALIGN,
        //
        visibleItemCount = pick(spec.visibleItemCount, data.length),
        nonHidden_visibleItemCount = 0;

    if (!spec.isPreRender && !isSizeLegend) {
        var _temp = 0;
        for (var p = 0; p < visibleItemCount; p++) {
            if (spec.data[p] && !spec.data[p].hideInLegend) {
                _temp++;
            }
            if (p === visibleItemCount - 1) {
                nonHidden_visibleItemCount = _temp;
            }
        }
    }
    spec.nonHidden_visibleItemCount = !spec.isPreRender && !isSizeLegend ? nonHidden_visibleItemCount : spec.nonHidden_visibleItemCount; //so prerendering the nonHidden_visibleItemCount will be got from exopndable-helpers itself
    //visibleItemCount - legend items covered in the area given (both hidden and non hidden items)
    //visibleCount - rendering item count - if all non hidden elements got covered in visibleitemcount, then all data(spec.data) will be rendered, else visibleItemCount will be the visibleCount.
    let visibleCount = !expandableConfig._expanded
            ? !spec.isPreRender && !isSizeLegend //we are looping each elemeent in expandable-helpers.js, so we can't render all elements all of a sudden.
                ? mathMax(0, legData_length - spec.nonHidden_visibleItemCount) > 0
                    ? visibleItemCount
                    : data.length
                : visibleItemCount
            : data.length,
        getViewData = visibleCount => {
            let d = config.reversed ? d3_values(data).reverse() : data;
            return defined(visibleCount) ? d.slice(0, visibleCount) : d;
        },
        viewData = getViewData(visibleCount),
        size = viewScale && mathMax(viewScale(viewData[viewData.length - 1]), viewScale(viewData[0])),
        renderList = () => {
            let newViewSpec = {
                ...viewSpec,
                bound: { width: viewBound.width, height: mathMin(maxBoxHeight, viewBound.height) },
                data: getViewData()
            };
            let viewObj = new ListView(newViewSpec);
            viewObj.create();
            viewObj.draw(self.encode.getElement());
        },
        keepTitleAligned = titleConfig && titleConfig.show && titleConfig.keepAligned && legendAlign.layout === HORIZONTAL_ALIGN,
        setDimension = ele => {
            let expanded = expandableConfig.show ? expandableState(self.encode.children[2]) : true;
            setLegendDimension(
                ele,
                legendAlign,
                keepTitleAligned,
                bound,
                config,
                expanded,
                helpers,
                data,
                spec,
                globalFont,
                expandableSpec
            );
        },
        { marker: boxConfig, label: labelConfig } = listConfig,
        _f = +utils.getVal(getValByPriority([listConfig.label, globalFont, $ZCG], 'fontSize')),
        boxSize = getBoxSize(boxConfig, labelConfig, legendAlign, viewScale, _f),
        maxBoxHeight = viewScale ? d3_max(viewData, d => boxSize.height(d)) : viewBound && viewBound.height;
    // spec.visibleCount = visibleCount; //the hidden legends will be included in this count

    var unitTestObj = chartObj && chartObj.legend.unitTestObject['zc-legend'].clr[spec.order];
    let backgroundSpec = { config },
        titleSpec = {
            align: legendAlign,
            config,
            globalFont,
            encode: spec.encode,
            computedInfo: spec.computedInfo && spec.computedInfo.title,
            onEnd: spec.titleOnEnd
        },
        viewSpec = {
            bound: scales.size && (size ? { width: size, height: maxBoxHeight } : viewBound),
            config,
            format: format && ((d, i) => format(labelData ? labelData[i] : d, i)),
            globalFont,
            data: viewData,
            ticks: ticks,
            scale: viewScale,
            scales,
            align: legendAlign,
            label: { fontSize: +utils.getVal(getValByPriority([listConfig.label, globalFont, $ZCG], 'fontSize')) },
            class: 'zc-legend-',
            computedInfo: spec.computedInfo && spec.computedInfo.legend,
            onEnd: spec.legSeriesOnEnd,
            unitTestObj: chartObj && chartObj.legend && chartObj.legend.unitTestObject['zc-legend'].clr[spec.order]
        },
        expandableSpec = {
            show: true,
            config: expandableConfig,
            globalFont,
            callbackArgs,
            renderList,
            bound,
            priorityArray: [config, globalFont, $ZCG],
            height: () => {
                let headerHeight = getBoundingRect(getEle(self, 'LegendTitleHtml')).height;
                if (keepTitleAligned) {
                    return mathMax(headerHeight, mathMin(maxBoxHeight, pick(size, viewBound.height)));
                } else {
                    return mathMin(maxBoxHeight, pick(size, viewBound.height));
                }
            },
            isHorizontal,
            padding: config.itemMargin,
            expanded: expandableConfig._expanded,
            remainingCount: isSizeLegend ? spec.data.length - visibleItemCount : legData_length - spec.nonHidden_visibleItemCount,
            margin: () => {
                if (keepTitleAligned && !expandableConfig.keepAligned) {
                    let headerWidth = getBoundingRect(getEle(self, 'LegendTitleHtml')).width;
                    if (unitTestObj) {
                        unitTestObj.expandable = {
                            ...unitTestObj.expandable,
                            'margin-left': headerWidth + 'px'
                        };
                    }
                    return `0 0 0 ${headerWidth}`;
                } else {
                    return 0;
                }
            },
            order: spec.order,
            setLegendDimension: () => setDimension(self.encode.getElement()),
            alignLegend: spec.computedInfo && spec.computedInfo.expandable && spec.computedInfo.expandable.alignLegend
        };

    return {
        titleSpec,
        backgroundSpec,
        viewSpec,
        expandableSpec,
        keepTitleAligned,
        keepExpandableAligned: expandableConfig.keepAligned, // TODO: work
        setDimension,
        initLegendComp: computedInfo && computedInfo.initLegendComp
    };
}
export function collectLegendDimVals(wrap, legendAlign, keepTitleAligned, bound, config, helpers, data, spec, globalFont) {
    let chart = spec.chartObj,
        header = wrap.selectAll(`ul.${LEGSERIES_HEADER_CLASSNAME}`),
        expandable = wrap.selectAll(`${DIV_ELEMENT}.${EXPANDABLE_CLASSNAME}`),
        { title: titleConfig, border: borderConfig = {} } = config,
        { expandable: expandableConfig = {} } = config.list,
        borderWidth = borderConfig.show ? borderConfig.size : 0,
        { maxWidth, maxHeight } = bound,
        //
        keepExpandableAligned = expandableConfig.show && expandableConfig.keepAligned,
        isColumnView = legendAlign.layout === 'horizontal',
        wrapMaxWidth = maxWidth,
        isSizeLegend = spec.encode === 'z',
        legData = !isSizeLegend && chart ? legend_helpers.getLegendData(chart, spec.order) : data,
        legData_length = !isSizeLegend && chart ? legend_helpers.getLegendDataLength(chart, legData) : data.length;

    if (spec.isPreRender) {
        //we roundoff while prerendering since, the height and width got from boundingclient is roundedoff value only, so we roundoff the maxHeight and width in expandable code, so to compensate we do it here also.
        maxWidth = mathRound(maxWidth);
        maxHeight = mathRound(maxHeight);
    }

    if (isColumnView && chart) {
        let extraSpaces = +config.marginLeft + +config.marginRight, // + (config.border.show ? 2 * config.border.size : 0),
            allowedLegMaxWidth = chart.canvasarea.width - extraSpaces;

        if (maxWidth > allowedLegMaxWidth) {
            maxWidth = allowedLegMaxWidth;
        }
        wrapMaxWidth = maxWidth;
    }
    if (!isSizeLegend) {
        wrap.style('max-width', `${maxWidth}px`);
    }
    //in future both size legend and discrete legend must have same logic
    maxWidth -= 2 * borderWidth;
    maxHeight -= 2 * borderWidth;
    if (isSizeLegend) {
        //we do specifically for sizelegend, inorder to avoid image diffs in size legend
        wrap.style('max-width', `${maxWidth}px`);
    }

    var unitTestObj = chart && chart.legend.unitTestObject['zc-legend'].clr[spec.order];

    var tempDimension = helpers && helpers.getItemDimension(data[0]),
        tempLabelHeight = tempDimension && tempDimension.label.height,
        tempItemHeight = tempDimension && tempDimension.item.height,
        maxItemHeight;

    // if (config.colorBox) {
    var fontSize = +utils.getVal(config.fontSize || globalFont.fontSize),
        colorBoxSize = legend_helpers.getFilterBoxSize(config.colorBox || {}, fontSize),
        colorBoxHeight = colorBoxSize[1];
    maxItemHeight = pick(config.itemHeight && +config.itemHeight, mathMax(tempLabelHeight, colorBoxHeight) + 2 * +config.itemMargin);
    maxItemHeight = mathMax(maxItemHeight, tempItemHeight);
    //the tempHeight calculated will not be proper incase of showAs, since the items will not be aligned due to inline-block, if vertical-align set to 'top' it would work fine, inorder to avoid image diff-avoided it
    // }

    if (chart) {
        // set if legend is scrollable
        chart.legend.hasScroll = legData_length * maxItemHeight > maxHeight;
    }

    // Max width for title
    let defaultTitleMaxWidth = keepTitleAligned ? LEGEND_TITLE_MAX_WIDTH_HORIZONTAL : LEGEND_TITLE_MAX_WIDTH_VERTICAL,
        titleMaxWidth = mathMin(perToPx(titleConfig.maxWidth || defaultTitleMaxWidth, maxWidth), maxWidth);
    header.style('max-width', `${titleMaxWidth < 0 ? 0 : titleMaxWidth}px`);
    if (unitTestObj) {
        unitTestObj.title_grp.maxWidth = titleMaxWidth;
    }

    //the values got in getBoundingRect is roundeoff only
    let headerDimension = getBoundingRect(header),
        expandableDimension = getBoundingRect(expandable);

    if (!keepTitleAligned /* || bound.adjust*/) {
        maxHeight = mathMin(maxHeight - headerDimension.height, maxHeight);
    }

    let maxW = maxWidth - (keepTitleAligned ? headerDimension.width : 0) - (keepExpandableAligned ? expandableDimension.width : 0);
    maxW -= isSizeLegend && (keepTitleAligned || keepExpandableAligned) ? 2 : 0;
    // let maxH = maxHeight - (!keepExpandableAligned ? expandableDimension.height : 0);

    if (expandableConfig.show) {
        expandable.styles({
            'max-width': `${maxW}px`
        });
    }

    if (unitTestObj && expandable.node() && expandableConfig.show) {
        unitTestObj.expandable = {
            ...unitTestObj.expandable,
            maxWidth: maxW
        };
    }

    return {
        wrapMaxWidth,
        maxW,
        expandableDimension,
        tempItemHeight,
        isSizeLegend,
        maxItemHeight,
        maxHeight
    };
}

function setLegendDimension(
    wrap,
    legendAlign,
    keepTitleAligned,
    bound,
    config,
    legendExpanded,
    helpers,
    data,
    spec,
    globalFont,
    expandableSpec
) {
    var isSizeLegend = spec.encode === 'z';
    if (isSizeLegend) {
        setDimensionForSizeLegend(
            wrap,
            legendAlign,
            keepTitleAligned,
            bound,
            config,
            legendExpanded,
            helpers,
            data,
            spec,
            globalFont,
            expandableSpec
        );
    } else {
        setDimensionForDiscreteLegend(
            wrap,
            legendAlign,
            keepTitleAligned,
            bound,
            config,
            legendExpanded,
            helpers,
            data,
            spec,
            globalFont,
            expandableSpec
        );
    }
}
function setDimensionForDiscreteLegend(
    wrap,
    legendAlign,
    keepTitleAligned,
    bound,
    config,
    legendExpanded,
    helpers,
    data,
    spec,
    globalFont,
    expandableSpec
) {
    let chart = spec.chartObj,
        legseries = wrap.selectAll(`ul.${LEGSERIES_LEGEND_CLASSNAME}`),
        series = wrap.selectAll(`li.${SERIES_LEGEND_CLASSNAME}`),
        header = wrap.selectAll(`ul.${LEGSERIES_HEADER_CLASSNAME}`),
        expandable = wrap.selectAll(`${DIV_ELEMENT}.${EXPANDABLE_CLASSNAME}`),
        scrollbarwid = 20,
        scrollType = legendAlign.layout,
        { title: titleConfig, border: borderConfig = {} } = config,
        { expandable: expandableConfig = {} } = config.list,
        borderWidth = borderConfig.show ? borderConfig.size : 0,
        { maxWidth, maxHeight } = bound,
        //
        keepExpandableAligned = expandableConfig.show && expandableConfig.keepAligned,
        keepAligned = getDisplay(keepTitleAligned, keepExpandableAligned),
        isSizeLegend = spec.encode === 'z',
        legData = !isSizeLegend && chart ? legend_helpers.getLegendData(chart, spec.order) : data,
        legData_length = !isSizeLegend && chart ? legend_helpers.getLegendDataLength(chart, legData) : data.length;
    var unitTestObj = chart && chart.legend.unitTestObject['zc-legend'].clr[spec.order];
    let _r = collectLegendDimVals(wrap, legendAlign, keepTitleAligned, bound, config, helpers, data, spec, globalFont);

    let { wrapMaxWidth, maxW, expandableDimension, tempItemHeight, maxItemHeight } = _r;
    maxHeight = _r.maxHeight;

    let maxH = maxHeight;
    //in on prerendering cases (!spec.isPreRender) - if legData_length === spec.visibleItemCount -> then all data would be rendered - so expandable will not be there
    if (data.length != spec.visibleItemCount && !isSizeLegend && expandable.node()) {
        var _h = bound.maxHeight,
            _w = bound.maxWidth;
        let expandableAdjust = ExpandableAdjustFactor(expandableConfig, _w, _h, globalFont, legendAlign, config.itemMargin);
        maxH -= $Browser.PHANTOMJS ? expandableDimension.height : mathMax(expandableAdjust[1], expandableDimension.height);
        maxH = mathMin(_h ? _h - expandableAdjust[1] : maxH, maxH);
    }
    // if (expandableConfig.show && isColumnView) {
    //     maxH = tempItemHeight ? mathMax(tempItemHeight, maxH) : maxH;
    // }

    //when maxW and maxH goes less than 0
    maxW = maxW < 0 ? 0 : maxW;
    maxH = tempItemHeight ? mathMax(tempItemHeight, maxH) : maxH;

    legseries.styles({
        'max-width': `${maxW}px`,
        'max-height': `${maxH}px`
    });

    if (unitTestObj) {
        unitTestObj.legend_grp = {
            ...unitTestObj.legend_grp,
            maxWidth: maxW,
            maxHeight: maxH
        };
    }

    var showAs = config.showAs,
        isAutoItemWidth = config.itemWidth === 'auto';

    if (showAs && isAutoItemWidth) {
        if (defined(spec.legendLabelMaxWidth)) {
            //spec.legendlabelMaxWidth will be got from title grp onend function
            if (unitTestObj) {
                unitTestObj.legend_grp.legend_nodes.label.width = spec.legendLabelMaxWidth;
            }
            wrap.selectAll('.zc-legend-legendlabel').style('width', spec.legendLabelMaxWidth + PX);
        } else {
            //for precalculation purpose.
            let maxItemWidth = helpers.getMaxItemWidth(data),
                constrainedMaxWidth = mathMin(maxItemWidth, maxW);
            // spec.itemMaxWidth = constrainedMaxWidth;

            let fontSize = +utils.getVal(config.fontSize || globalFont.fontSize),
                colorBoxSize = legend_helpers.getFilterBoxSize(config.list.marker, fontSize);
            // #ZC1904 subtract the itemMargin
            var legendLabelMaxWidth = constrainedMaxWidth - colorBoxSize[0] - config.itemMargin * 2;
            if (defined(legendLabelMaxWidth)) {
                series.selectAll('.zc-legend-legendlabel').style('width', legendLabelMaxWidth + PX);
                if (unitTestObj) {
                    unitTestObj.legend_grp.legend_nodes.label.width = legendLabelMaxWidth;
                }
            }
        }
        if ($Browser.WINDOWS && scrollType === HORIZONTAL_ALIGN) {
            let adjustMaxWidth = 2;
            legseries.style('max-width', `${maxW + adjustMaxWidth}px`);
            if (unitTestObj) {
                unitTestObj.legend_grp.maxWidth = maxW + adjustMaxWidth;
            }
        }
    } else {
        series.selectAll('.zc-legend-legendlabel').style('width', null);
    }
    //TODO make the code common for size legend and discrete legend
    // +2 added in many places, need to check it. lot of diffs if its removed // TODO Check

    legseries.styles({
        'max-width': `${maxW}px`,
        'max-height': `${maxH}px`
    });
    if (scrollType === HORIZONTAL_ALIGN) {
        legseries.styles({ width: NULL, height: NULL });
        series.style('max-width', `${maxW}px`);

        if (unitTestObj) {
            unitTestObj.legend_grp.legend_nodes.maxWidth = maxW;
        }

        var headerDisplay = header.style('display');
        wrap.style('max-width', `${maxW + 2 * borderWidth}px`);
        if (!spec.isPreRender) {
            if (maxItemHeight) {
                header.styles({ display: 'block' });
                //setting only legseries inline, will not give proper bound width
                legseries.styles({ display: 'inline', 'border-top': '1px solid' });
            } else {
                //maxItemHeight will not be there custom legend case, since precalculation will not be done for it, so for that we keep display as 'inline' only then the boundbox height will be got properly, border - inorder to avoid image diff
                legseries.styles({ display: null, border: '1px solid' });
            }
        }

        let bb = getBoundingRect(legseries);
        header.styles({ display: headerDisplay });
        legseries.styles({ display: titleConfig.keepAligned ? 'inline-block' : null, border: NULL, 'border-top': NULL });

        wrap.style('max-width', `${wrapMaxWidth}px`);
        //in case of prerender, all data will be rendered, in the visiblecount hidden elemeents also will be included, but in actual renmdering hidden elements will not be included,
        let scrollWid =
            elementIsScrollable(legseries) || (expandableConfig.show && spec.nonHidden_visibleItemCount < legData_length)
                ? scrollbarwid
                : 0;
        legseries.style('width', `${bb.width + scrollWid + 2}px`);
        series.style('max-width', `${maxW - scrollWid > 0.5 * (maxW) ? maxW - scrollWid : maxW}px`);

        var activeRows;
        if (!spec.isPreRender) {
            if (spec.rowCount) {
                activeRows = spec.rowCount;
            } else {
                activeRows = mathRound(legseries.node().offsetHeight / maxItemHeight) || 1;
            }
        }
        //we can't get the height from bb sice whne inline is applied to legeseries, the height got will be 0 and can't be used, so we use activeROws instead to set the height
        //actually we need to floor the value, since only then no half legend item will be got, but we are roundingoff since when item size is 22.5, we get maxItemHeight as 23 but legseries.node().offsetHeight will be properly based on 22.5,
        //so now even though 6 rows can be there, since .5 is extra for each case instead of getting 6.1.. we will get 5.8 something so, if we floor only 5 rows will come, but actually 6 is possible, so we round instead of floor
        scrollWid = elementIsScrollable(legseries, 'width') ? scrollbarwid : 0;
        legseries.style('height', `${activeRows * maxItemHeight + scrollWid + 2}px`);

        if (spec.isPreRender) {
            legseries.style('height', `${bb.height + scrollWid}px`);
            //in actual rendering bb.height will not be proper due to inline property
        }
        if (!maxItemHeight) {
            legseries.style('height', `${bb.height + scrollWid + 2}px`);
        }

        if (unitTestObj) {
            unitTestObj.legend_grp = {
                ...unitTestObj.legend_grp,
                preCalculationStyle: {
                    width: null,
                    height: null,
                    display: 'inline',
                    'border-top': '1px solid'
                },
                width: bb.width + scrollWid + 2,
                height: activeRows * maxItemHeight + scrollWid + 2,
                'border-top': null,
                display: titleConfig.keepAligned ? 'inline-block' : null
            };
        }
    } else {
        if (spec.nonHidden_visibleItemCount < legData_length) {
            //allocate space only for the visible items within range
            maxH = mathMin(spec.nonHidden_visibleItemCount * tempItemHeight, maxH);
        }

        legseries.styles({
            'max-height': `${maxH}px`
        });

        let scrollWid =
            elementIsScrollable(legseries) || (expandableConfig.show && spec.nonHidden_visibleItemCount < legData_length)
                ? scrollbarwid
                : 0;
        // reset the width first
        legseries.styles({ width: NULL /*, height: NULL*/ });

        // calculate the max item width
        let legendWidth = getBoundingRect(legseries.style('display', 'inline-block')).width;
        legseries.style('display', keepAligned);
        legseries.style('width', `${legendWidth + scrollWid + 2}px`);
        legseries.style('height', NULL);

        // if (!spec.isPreRender) {
        //below for legend4, since on expanding size of the legend got varied
        // var activeRows = mathFloor(legseries.node().offsetHeight / tempItemHeight) || 1;
        // scrollWid = elementIsScrollable(legseries, 'width') ? scrollbarwid : 0;
        // legseries.style('height', `${activeRows * tempItemHeight + scrollWid}px`);
        // }

        if (unitTestObj) {
            unitTestObj.legend_grp = {
                ...unitTestObj.legend_grp,
                maxHeight: maxH,
                preCalculationStyle: {
                    width: null,
                    display: 'inline-block'
                },
                width: legendWidth + scrollWid + 2,
                height: null,
                display: keepAligned
            };
        }
    }
    var addOverFlowY = ((expandableConfig.show && legendExpanded) || !expandableConfig.show) === true;
    if ($Browser.IE && addOverFlowY) {
        addOverFlowY = legseries.node().scrollHeight - spec.visibleItemCount > legseries.node().offsetHeight;
    }
    spec.addOverFlowY = addOverFlowY;
    if (unitTestObj) {
        unitTestObj.legend_grp.addOverFlowY = addOverFlowY;
    }
    var addOverFlowX = elementIsScrollable(legseries, 'width') ? scrollbarwid : 0;
    legseries.style('overflow-x', addOverFlowX ? 'auto' : 'hidden');
    legseries.style('overflow-y', addOverFlowY ? 'auto' : 'hidden');

    if (unitTestObj) {
        unitTestObj.maxWidth = wrapMaxWidth;
    }
    if (!spec.isPreRender) {
        if (expandableSpec) {
            expandable.styles({
                margin: `${expandableSpec.margin()}px`
            });
        }
    }
}
function setDimensionForSizeLegend(
    wrap,
    legendAlign,
    keepTitleAligned,
    bound,
    config,
    legendExpanded,
    helpers,
    data,
    spec,
    globalFont,
    expandableSpec
) {
    let legseries = wrap.selectAll(`ul.${LEGSERIES_LEGEND_CLASSNAME}`),
        series = wrap.selectAll(`li.${SERIES_LEGEND_CLASSNAME}`),
        header = wrap.selectAll(`ul.${LEGSERIES_HEADER_CLASSNAME}`),
        expandable = wrap.selectAll(`${DIV_ELEMENT}.${EXPANDABLE_CLASSNAME}`),
        scrollbarwid = 20,
        scrollType = legendAlign.layout,
        { title: titleConfig, border: borderConfig = {} } = config,
        { expandable: expandableConfig = {} } = config.list,
        borderWidth = borderConfig.show ? borderConfig.size : 0,
        { maxWidth, maxHeight } = bound,
        //
        keepExpandableAligned = expandableConfig.show && expandableConfig.keepAligned,
        keepAligned = getDisplay(keepTitleAligned, keepExpandableAligned);

    maxWidth -= 2 * borderWidth;
    maxHeight -= 2 * borderWidth;

    // Max width for title
    let defaultTitleMaxWidth = keepTitleAligned ? LEGEND_TITLE_MAX_WIDTH_HORIZONTAL : LEGEND_TITLE_MAX_WIDTH_VERTICAL,
        titleMaxWidth = mathMin(perToPx(titleConfig.maxWidth || defaultTitleMaxWidth, maxWidth), maxWidth);
    header.style('max-width', `${titleMaxWidth}px`);

    let headerDimension = getBoundingRect(header),
        expandableDimension = getBoundingRect(expandable);

    if (!keepTitleAligned) {
        maxHeight = mathMin(maxHeight - headerDimension.height, maxHeight);
    }

    wrap.style('max-width', `${maxWidth}px`);
    legseries.styles({
        'max-width': `${maxWidth}px`,
        'max-height': `${maxHeight}px`
    });

    if (scrollType === HORIZONTAL_ALIGN) {
        legseries.styles({ width: NULL, height: NULL });

        let maxW = maxWidth - (keepTitleAligned ? headerDimension.width : 0) - (keepExpandableAligned ? expandableDimension.width : 0);
        maxW -= keepTitleAligned || keepExpandableAligned ? 2 : 0;
        let maxH = maxHeight - (!keepExpandableAligned ? expandableDimension.height : 0);

        legseries.style('max-width', `${maxW}px`);
        legseries.style('max-height', `${maxH}px`);

        legseries.style('border', '1px solid');
        let bb = getBoundingRect(legseries);
        legseries.style('border', NULL);

        legseries.style('width', `${bb.width + 2}px`);
        let scrollWid = elementIsScrollable(legseries, 'width') ? scrollbarwid : 0;
        legseries.style('height', `${bb.height + scrollWid + 2}px`);

        let addOverFlowY = legseries.node().scrollHeight - 2 > legseries.node().offsetHeight;
        legseries.style('overflow-x', 'auto');
        legseries.style('overflow-y', addOverFlowY ? 'auto' : 'hidden');
    } else if (legendAlign.layout === HORIZONTAL_ALIGN) {
        // horizontal align ------->

        // reset the dimension for auto calculation, if any
        legseries.styles({ width: NULL, height: NULL });

        // set the max width for each item. it will solve text overlapping issue.
        series.style('max-width', `${maxWidth}px`);

        // set the calculated max width in the wrap element,
        // so that when we set inline to the legseries[ul], it won't stress to the full width.
        // will reset after calculating the inline content width
        wrap.style('max-width', `${maxWidth}px`);

        // calculate the actual content width
        // previously we were calculating with each item and sum their width.

        // apply dummy style to get the content width
        // the border will set the max with to the parent
        legseries.styles({ display: 'inline', 'border-top': '1px solid' });

        let legendWidth = getBoundingRect(legseries).width;
        legseries.styles({ display: keepAligned, 'border-top': NULL });

        wrap.style('max-width', `${maxWidth}px`); // resetting it

        // do need the below 2 lines. to get rid of image diff setting the height.
        legseries.style('height', `${getComputedHeight(series, maxWidth, legendAlign) + 2}px`);

        // check the scroll width after setting the height; // phantomjs issue
        let scrollWid = elementIsScrollable(legseries) ? scrollbarwid : 0;

        // inline block needs and width. otherwise when it auto wraps, it sets the width to the max-width
        legseries.style('width', `${legendWidth + scrollWid + 2}px`);
        legseries.style('overflow', 'auto');
    } else {
        // if ($Browser.IE && addOverFlowY) {
        //     addOverFlowY = legseries.node().scrollHeight - activeRows > legseries.node().offsetHeight;
        // }

        let maxH = maxHeight - expandableDimension.height;
        legseries.style('max-height', `${maxH}px`);

        let scrollWid = elementIsScrollable(legseries) ? scrollbarwid : 0;
        // reset the width first
        legseries.style('width', NULL);

        // calculate the max item width
        let legendWidth = getBoundingRect(legseries.style('display', 'inline-block')).width;
        legseries.style('display', keepAligned);
        legseries.style('width', `${legendWidth + scrollWid + 2}px`);
        legseries.style('height', NULL);

        //  ZC2423 - if expandable, overflow-x & overflow-y needs to be updated for IE
        legseries.style('overflow-x', legendExpanded ? 'auto' : 'hidden');
        legseries.style('overflow-y', legendExpanded ? 'auto' : 'hidden');
    }
}

function getComputedHeight(series, maxWidth) {
    let h = 0,
        prevWidth = 0,
        prevHeight = 0;

    series.each(function() {
        let bb = getBoundingRect(this);
        if (prevWidth + bb.width > maxWidth) {
            h += prevHeight;
            prevHeight = 0;
            prevWidth = 0;
        }

        prevWidth += bb.width;
        prevHeight = mathMax(prevHeight, bb.height);
    });
    h += prevHeight;

    return h;
}

function getDisplay(keepTitleAligned, keepExpandableAligned) {
    return keepTitleAligned || keepExpandableAligned ? 'inline-block' : NULL;
}

function getEle(self, view) {
    let order = {
        LegendTitleHtml: 0
    };
    return self.encode.children[0].children[order[view]].getElement();
}

export { ListLegendParser };
