// @zohocharts/zc/src/components/race/intermediateFrames/defaultFrameDataInterpolator.js
import { isUndefined, simpleClone } from '../../../utils/zcutils';
import { convertDataInto3D, getDataIndicesToBeInterpolated, getDataKeys, interpolateDataPoint } from './helpers';

export function defaultframeInterpolator(race, frameInfo) {
    this.race = race;
    this.frameInfo = frameInfo;
}


defaultframeInterpolator.prototype.getFramesWithDataAccessor = framesWithDataAccessor;
defaultframeInterpolator.prototype.getFramesWithoutDataAccessor = framesWithDataAccessor;

defaultframeInterpolator.prototype.getFrames = function(){
    var frameInfo = this.frameInfo,
        race = this.race;
        var { fromFrame, toFrame, intermediateFramesCount, racingColumns, customInterpolator } = frameInfo,
        intermediateFrames = [],
        chart = race.chart;

    convertDataInto3D(fromFrame, toFrame, chart);
    var yaxiscolumnorder = fromFrame.yaxiscolumnorder,
        xColInfo = chart.dataset.getColumn('x'),
        yColInfo = chart.dataset.getColumn('y', yaxiscolumnorder || [0, 0]),
        ydatatype = yColInfo.datatype,
        xDataType = xColInfo.datatype,
        isBothXYOrdinal = xDataType === 'ordinal' && ydatatype === 'ordinal',
        isBothXYNotOrdinal = xDataType != 'ordinal' && ydatatype != 'ordinal';

    if (isBothXYOrdinal || isBothXYNotOrdinal) {
        framesWithoutDataAccessor(racingColumns, fromFrame, toFrame, intermediateFrames, intermediateFramesCount, chart);
    } else {
        var xDataIndex = chart.dataset.getColumn('x').dataindex,
            yDataIndex = chart.dataset.getColumn('y', yaxiscolumnorder || [0, 0]).dataindex,
            uniqueDataKeyIndex = xDataType === 'ordinal' ? xDataIndex : yDataIndex;
        framesWithDataAccessor(uniqueDataKeyIndex, racingColumns, fromFrame, toFrame, intermediateFrames, intermediateFramesCount, chart);
    }
    intermediateFrames.shift(); //from frmae will be mepty, so we remove it in here.
    return intermediateFrames;
}

function framesWithDataAccessor(uniqueDataKeyIndex, racingColumns, fromFrame, toFrame, intermediateFrames, intermediateFramesCount, chart) {
    var fromFrameData = fromFrame.data[0],
        toFrameData = toFrame.data[0],
        to_Category_DataMap = {},
        fromKeys = fromFrame.keys || getDataKeys(fromFrameData, uniqueDataKeyIndex),
        toKeys = toFrame.keys || getDataKeys(toFrameData, uniqueDataKeyIndex),
        enterKeys = toKeys.filter(o => fromKeys.indexOf(o) === -1), //new categories
        enterData = [],
        exitData = [];
    for (var i = 0; i < toFrameData.length; i++) {
        to_Category_DataMap[toFrameData[i][uniqueDataKeyIndex]] = toFrameData[i];
    } //creating a map inorder to avoid loop of toData each time for calculating the interpolation

    for (var i = 0; i < fromFrameData.length; i++) {
        var _fromdata, _todata;
        var _fromcat = fromFrameData[i][uniqueDataKeyIndex]; //this has to be the unique value of each datapoint based on which race occures..now considering x as that value.
        _fromdata = fromFrameData[i];
        _todata = to_Category_DataMap[_fromcat];

        var interpolationIndices = getDataIndicesToBeInterpolated(racingColumns, chart);
        interpolateDataPoint(_todata ? intermediateFramesCount : 0, interpolationIndices, _fromdata, _todata, intermediateFrames);
        if(isUndefined(_todata)){
            exitData.push(simpleClone(_fromdata));
        }
    }
    if (enterKeys.length) {
        for (var i = 0; i < enterKeys.length; i++) {
            var data = to_Category_DataMap[enterKeys[i]];
            enterData.push(simpleClone(data));
        }
    }
    return {
        intermediateFrames,
        enterData,
        exitData
    }
    
}

function framesWithoutDataAccessor(racingColumns, fromFrame, toFrame, intermediateFrames, intermediateFramesCount, chart) {
    var fromFrameData = fromFrame.data[0],
        toFrameData = toFrame.data[0];

    //direct array mapping, no reordering, interpolation happens 0 to 0, 1 to 1, based on the input given.
    for (var i = 0; i < fromFrameData.length; i++) {
        var _fromdata, _todata;
        _fromdata = fromFrameData[i];
        _todata = toFrameData[i];

        var interpolationIndices = getDataIndicesToBeInterpolated(racingColumns, chart);
        interpolateDataPoint(_todata ? intermediateFramesCount : 0, interpolationIndices, _fromdata, _todata, intermediateFrames);
    }
    return { intermediateFrames };
}
