// @zohocharts/zc/src/components/race/intermediateFrames/helpers.js
import { d3_scaleLinear, d3_scaleOrdinal, d3_scaleTime } from '../../../core/d3-constants';
import { DataProcessor } from '../../../core/data-processor';
import { NUMERIC_DATATYPE } from '../../../core/datatype-constants';
import { generateDateVal } from '../../../formatter/formatter';
import { isArray, isUndefined, simpleClone } from '../../../utils/zcutils';
export function convertDataInto3D(fromFrame, toFrame, chart) {
    var processor = chart.dataObject,
        helpers = DataProcessor.helpers,
        metadata = processor.data.metadata,
        groupbyDataIndices = helpers.groupByDataIndices(metadata, chart);

    if (!helpers.is3D(fromFrame.data) && isArray(fromFrame.data)) {
        var chartType_num = helpers.getChartType(fromFrame.type || processor.globalChartType, NUMERIC_DATATYPE);
        fromFrame.data = processor.convertTo3DArray(fromFrame, groupbyDataIndices, chartType_num);
    }
    if (!helpers.is3D(toFrame.data) && isArray(toFrame.data)) {
        var chartType_num = helpers.getChartType(toFrame.type || processor.globalChartType, NUMERIC_DATATYPE);
        toFrame.data = processor.convertTo3DArray(toFrame, groupbyDataIndices, chartType_num);
    }
}

export function getDataIndicesToBeInterpolated(racingColumns, chart) {
    var indices = [];
    racingColumns.map(function(d, i) {
        indices.push(chart.systemConf.metadata.columns[d].dataindex);
    });
    return indices;
}

export function interpolateDataPoint(count, indicesToInterpolate, fromDataPoint, toDataPoint, intermediateFrames) {
    var interpolator_map = d3.map();
    var isExitFrame = isUndefined(toDataPoint);
    if (!isExitFrame) {
        indicesToInterpolate.map(function(d) {
            interpolator_map.set(d, d3.interpolate(fromDataPoint[d], toDataPoint[d]));
        });
    }
    var data = simpleClone(fromDataPoint);

    for (var i = 0; i < count; i++) {
        var t = i / (count - 1);

        var isFromFrame = i == 0,
            isToFrame = i == count - 1;
        if (!(isFromFrame || isToFrame)) {
            if (!intermediateFrames[i]) {
                intermediateFrames[i] = [[]];
            }
            indicesToInterpolate.map(function(d) {
                data[d] = interpolator_map.get(d)(t);
            });
            intermediateFrames[i][0].push(simpleClone(data));
        }
    }
}

export function getDataKeys(data, uniqueDataKeyIndex) {
    var keys = [];
    data.map(function(d, i) {
        keys.push(d[uniqueDataKeyIndex]);
    });
    return keys;
}

export function getKeyFramePosScale(frames, datatype, intermediateFramesSettings, idMeta) {
    var chart = this.chart,
        scroll = chart.scroll,
        getFirstAndLastTimeFrameId = frames => {
            var first = frames[0],
                last = frames[frames.length - 1];

            return {
                first: first.keyFrame_Id || first[0].keyFrame_Id, //multiseries frames
                last: last.keyFrame_Id || last[0].keyFrame_Id
            };
        };

    if (scroll) {
        var raceScale,
            range = scroll.settings.orient
                ? scroll.settings.orient === 'right' || scroll.settings.orient === 'left'
                    ? [scroll.brushExtent[0][1], scroll.brushExtent[1][1]]
                    : [scroll.brushExtent[0][0], scroll.brushExtent[1][0]]
                : chart.axes.rotated
                ? [scroll.brushExtent[0][1], scroll.brushExtent[1][1]]
                : [scroll.brushExtent[0][0], scroll.brushExtent[1][0]];

        if(scroll.overviewEnabled){
            range = chart.overview.axes.x.scale.range();
        }
        switch (datatype) {
            case 'time':
                {
                    var { first, last } = getFirstAndLastTimeFrameId(frames);
                    (first = generateDateVal(first, null, idMeta[datatype])), (last = generateDateVal(last, null, idMeta[datatype]));
                    var domain = [first, last];
                    raceScale = d3_scaleTime()
                        .domain(domain)
                        .range(range);
                }
                break;
            case 'numeric':
                {
                    var { first, last } = getFirstAndLastTimeFrameId(frames),
                        domain = [first, last];
                    raceScale = d3_scaleLinear()
                        .domain(domain)
                        .range(range);
                }
                break;
            case 'ordinal': {
                var totalFramesCount =
                    intermediateFramesSettings.count * (chart.userdata.keyframes.length - 1) + chart.userdata.keyframes.length;
                var domain = d3.range(totalFramesCount - 1);
                raceScale = d3_scaleOrdinal()
                    .domain(domain)
                    .range(range);
            }
        }
        return raceScale;
    }
    return null;
}

export function getKeyFrameIDInterpolator(keyFrameType, fromFrame, toFrame, meta, keyFramePosScale) {
    if (keyFrameType === 'time') {
        var from = generateDateVal(fromFrame.keyFrame_Id, null, meta[keyFrameType]),
            to = generateDateVal(toFrame.keyFrame_Id, null, meta[keyFrameType]);
        return d3.interpolate(from, to);
    } else if (keyFrameType === 'numeric') {
        return d3.interpolate(keyFramePosScale.domain()[0], keyFramePosScale.domain()[1]);
    }
}

export function getFramePositionByIndex(index, _framePosArgs, posArr = []) {
    var race = this,
        chart = race.chart,
        { keyFramePosScale, isUpdaterRacePosArr: updatePosArr, keyFrameType, keyframeidInterpolator, count } = _framePosArgs,
        isScrolledRace = chart.scroll;

    if (isScrolledRace) {
        var position_val;
        if (keyFrameType === 'ordinal') {
            position_val = keyFramePosScale(race.data.length > 0 ? race.data.length - 1 + index : index);
        } else {
            var currentKeyFrameId = keyframeidInterpolator(index / (count + 2 - 1));
            position_val = keyFramePosScale(currentKeyFrameId);
        }
        if (updatePosArr) {
            if (isUndefined(position_val)) {
                posArr.push(keyFramePosScale.range()[1]);
                //inorder to match ordinal with numeric and time, we doesnot consider last ordinal frame in scale, and put the value directly, to get the distribution properly.
                return posArr;
            }
            if (index === 0) {
                if (race.data.length === 0) {
                    posArr.push(position_val);
                }
            } else {
                posArr.push(position_val);
            }
        }
    }
    return posArr;
}
