// @zohocharts/zc/src/components/race/intermediateFrames/init.js
import { invokeFunction } from '../../../utils/zcutils';
import { defaultframeInterpolator } from './defaultFrameDataInterpolator';
import { convertDataInto3D, getFramePositionByIndex, getKeyFrameIDInterpolator, getKeyFramePosScale } from './helpers';

export function generateIntermediateKeyFrames(frames) {
    var race = this,
        raceSettings = race.userRaceSettings,
        keyFramesConf = raceSettings.frames,
        intermediateFramesSettings = keyFramesConf.intermediateFrames,
        keyFrameType = keyFramesConf.metadata.column.datatype,
        racingColumns = keyFramesConf.metadata.racingColumns,
        keyFramePosScale = getKeyFramePosScale.call(race, frames, keyFrameType, intermediateFramesSettings, keyFramesConf.metadata.column),
        frameIdx,
        frameInfo = {
            keyFramePosScale,
            keyFrameType,
            ...intermediateFramesSettings,
            intermediateFramesCount: intermediateFramesSettings.count + 2,
            frameIdx: 0,
            fromFrame: null,
            toFrame: null,
            racingColumns: racingColumns,
            meta: keyFramesConf.metadata.column
        },
        multiSeriesRace = !!frames[0].length;

    if (multiSeriesRace) {
        race.data = [];
        for (frameIdx = 0; frameIdx < frames.length - 1; frameIdx++) {
            var generatedFramesArr = [],
                seriesCount = frames[frameIdx].length;
            for (var seriesIdx = 0; seriesIdx < seriesCount; seriesIdx++) {

                frameInfo.fromFrame = frames[frameIdx][seriesIdx];
                frameInfo.toFrame = frames[frameIdx + 1][seriesIdx];
                frameInfo.frameIdx = frameIdx;

                
                var _frames = _framesGenerator.call(race, frameInfo, seriesIdx === 0 ? true : false);
                _frames.map(function(frame, i) {
                    if (!generatedFramesArr[i]) {
                        generatedFramesArr[i] = [];
                    }
                    generatedFramesArr[i].push(frame);
                });
            }
            race.data = race.data.concat(generatedFramesArr);
            //for ordinal we ignore last datapoint in keyFramePosScale, inorder to macth numeric and ordinal scale, so we add the last pos here.
            if (race.positionArray.length && race.positionArray[race.positionArray.length - 1] === undefined) {
                race.positionArray[race.positionArray.length - 1] = keyFramePosScale.range()[1];
            }
        }
    } else {
        race.data = [];
        for (frameIdx = 0; frameIdx < frames.length - 1; frameIdx++) {

            frameInfo.fromFrame = frames[frameIdx];
            frameInfo.toFrame = frames[frameIdx + 1];
            frameInfo.frameIdx = frameIdx;
            var _frames = _framesGenerator.call(race, frameInfo, true);
            race.data = race.data.concat(_frames);
            if (race.positionArray.length && race.positionArray[race.positionArray.length - 1] === undefined) {
                race.positionArray[race.positionArray.length - 1] = keyFramePosScale.range()[1];
            }
        }
    }
}

function _framesGenerator(frameInfo, updateFramePositionArray) {
    var { fromFrame, toFrame, racingColumns, customInterpolator } = frameInfo,
        intermediateFrames;

    if (customInterpolator) {
        convertDataInto3D(fromFrame, toFrame, this.chart);
        var args = {
            fromFrame,
            toFrame,
            racingColumns,
            frameInfo,
            defaultframeInterpolator
        };
        intermediateFrames = invokeFunction(customInterpolator, args, this);
        //pass default too here
    } else {
        var frameInterpolator = new defaultframeInterpolator(this, frameInfo);
        intermediateFrames = frameInterpolator.getFrames();
    }
    var _frames = constructIntermediateFrames.call(this, frameInfo, intermediateFrames, updateFramePositionArray);
    _frames = this.data.length === 0 ? [fromFrame].concat(_frames) : _frames;
    _frames.push(toFrame);
    return _frames;
}

function constructIntermediateFrames(frameInfo, intermediateFrames, isUpdaterRacePosArr) {
    var { fromFrame, toFrame, dataSort, _constructor, keyFramePosScale, keyFrameType, meta } = frameInfo,
        race = this,
        invokeFunction = $ZC.getComponent('utils.invokeFunction'),
        keyframeidInterpolator = getKeyFrameIDInterpolator(keyFrameType, fromFrame, toFrame, meta, keyFramePosScale),
        interFramePosArr = [],
        _framePosArgs = {
            keyFramePosScale,
            isUpdaterRacePosArr,
            keyFrameType,
            keyframeidInterpolator,
            count: intermediateFrames.length
        },
        dataSortArgs = {
            fromFrame,
            toFrame,
            intermediateFrames,
            keyFramePosScale,
            keyframeidInterpolator
        };

    if (!race.positionArray) {
        race.positionArray = [];
    }
    var frames = intermediateFrames.map(function(d, i) {
            //i+1, since we need only for intermediate frames, to avoid from frame we do +1 to i in keyFramePosScale
            var args = {
                    ...dataSortArgs,
                    overallFrameIdx: race.data.length - 1 + i
                },
                _frame = invokeFunction(dataSort, d, i, args);
            getFramePositionByIndex.call(race, i + 1, _framePosArgs, interFramePosArr);
            return invokeFunction(_constructor, _frame, i, args);
        }),
        fromFramePosition = getFramePositionByIndex.call(race, 0, _framePosArgs),
        toFramePosition = getFramePositionByIndex.call(race, intermediateFrames.length - 1 + 2, _framePosArgs);

    race.positionArray = race.positionArray.concat(fromFramePosition.concat(interFramePosArr).concat(toFramePosition));
    return frames;
}
