// @zohocharts/zc/src/dial/dialchart.js
//$Id$
//dialchart.js
import { svg_plot_renderer, labelRenderer } from '../core/main-zc-init';
import { mergeJSON, validateNullVal, defined, perToPx, pick } from '../utils/zcutils';
import { getAnimationConf } from '../utils/animationutils';
import { PATH_ELEMENT } from '../core/element-constants';
import { d3_arc, d3_ease_linear, d3_interpolate, d3_select, d3_descending } from '../core/d3-constants';
import { arcInnerTween } from '../pie/pieAnimationUtils';
import { getInnerRadius, radianToDegree } from '../polaraxis/polarutils';
import { dialUtils } from './dial-utils';
import { appendEle } from '../utils/zcsvgutils';
import { effects, getGradientProp } from '../effects/zceffects';
import { DIAL_CHARTNAME } from '../core/chartname-constants';
import { mathMax, mathMin } from '../core/zc-constants';
import { NONE } from '../core/zc-constants';
import { getSvgTransform } from '../utils/svg-transforms';
import { isFacet } from '../axis/hierarchy-axis-utils';
import { getColorByIndex } from '../utils/colorutils';
/**
 *
 * @returns {chart} dial chart renderer.
 */
/*
 * PENDING:-
 * If target label is overlapping with minmax label
 * Pre render target
 */
svg_plot_renderer.dial = function() {
    var commonRendererProp,
        chartObj,
        innerRadius,
        color,
        isredraw = false,
        commonPlotOptions,
        dataset,
        yaxiscolumnorder,
        totalEngagement,
        plotarea_approx,
        systemconf,
        yscale,
        oldData,
        minMax,
        hasNegValue,
        hasPreviousNegValue,
        previousRadiusFactors,
        previousOuterRadius,
        previousInnerRadius,
        currentRadiusFactors,
        arc,
        _data;
    var pie = svg_plot_renderer.pie(),
        valueAccessor = d => dataset.getY(d, yaxiscolumnorder);

    function chart(selection) {
        selection.each(function(data) {
            //renderer config values
            var rp = commonRendererProp.params,
                { bound } = rp;
            (chartObj = rp.chartObj),
                (commonPlotOptions = rp.commonPlotOptions),
                (color = rp.color),
                (dataset = rp.dataset),
                (yaxiscolumnorder = rp.yaxiscolumnorder);
            systemconf = chartObj.systemConf;
            _data = data[0].data[0];

            var plotoptions = data[0].plotoptions,
                serData = data[0],
                dataCount = _data.length,
                paneid = isFacet(chartObj) ? `${serData.row}_${serData.column}` : null;
            totalEngagement = valueAccessor(_data);
            hasPreviousNegValue = hasNegValue;
            hasNegValue = totalEngagement < 0;
            var defsLocation = chartObj.defsLocation;
            var labelQueue = rp.labelQueue,
                { rendererIndex } = rp,
                seriesIndex = 0;

            yscale = rp.yscale || chartObj.axes.y.groups[0].scale;
            minMax = yscale.domain();
            plotarea_approx = bound.polar;

            commonPlotOptions = plotoptions ? mergeJSON(commonPlotOptions, plotoptions, true) : commonPlotOptions;
            //Animation variables
            var animationOptions = getAnimationConf(commonPlotOptions, plotoptions, chartObj, DIAL_CHARTNAME);
            var isAnimationEnable = animationOptions.enabled,
                easingType = animationOptions.easingType,
                transitionDuration = isredraw ? (isAnimationEnable ? 200 : 0) : animationOptions.duration,
                _generatedData = createDataForPie(data);

            var container = d3_select(this).data(_generatedData);
            oldData = container.selectAll(`${PATH_ELEMENT}.dial`).data();

            var levelMarkerConf = commonPlotOptions.levelMarker;
            currentRadiusFactors = `${commonPlotOptions.innerRadius} _ ${commonPlotOptions.outerPadding} _ ${commonPlotOptions.maxRadius} _ ${commonPlotOptions.bandWidth} _ ${levelMarkerConf.bandWidth} _ ${levelMarkerConf.innerPadding} _ ${levelMarkerConf.outerPadding}`;
            var OR = plotarea_approx.radius,
                levelmarkerSW = validateNullVal(levelMarkerConf.strokeWidth, 1),
                visibility,
                cornerRadius = pick(levelMarkerConf.cornerRadius, commonPlotOptions.cornerRadius),
                padAngle = pick(levelMarkerConf.padAngle, commonPlotOptions.padAngle);

            var { mode = 'radial' } = commonPlotOptions;
            var IR = getInnerRadius(commonPlotOptions, OR);
            arc = d3_arc()
                .endAngle(function(d) {
                    return yscale(d[1]);
                })
                .cornerRadius(cornerRadius)
                .padAngle(padAngle);

            var levelsInfo = (chart.levelMarker = dialUtils.createLevelsNColors(
                    data,
                    levelMarkerConf,
                    getColorByIndex(chartObj, seriesIndex, paneid),
                    minMax
                )),
                levelmarkerClr = levelsInfo.colors;

            if (mode === 'radial') {
                var innerPadding, outerPadding;

                if (defined(levelMarkerConf.bandWidth) || defined(levelMarkerConf.maxBandWidth)) {
                    var maxLevelMarkerBandWidth = validateNullVal(levelMarkerConf.maxBandWidth, OR);
                    var levelMarkerBandWidth = Math.min(perToPx(levelMarkerConf.bandWidth, OR, OR - IR), maxLevelMarkerBandWidth);

                    (innerPadding = (OR - IR - levelMarkerBandWidth) / 2), (outerPadding = (OR - IR - levelMarkerBandWidth) / 2);

                    visibility = levelMarkerBandWidth === 0 ? 'hidden' : null;
                } else {
                    (innerPadding = mathMax(0, perToPx(levelMarkerConf.innerPadding, OR, 0))),
                        (outerPadding = mathMax(0, perToPx(levelMarkerConf.outerPadding, OR, 0)));
                }

                arc.innerRadius(IR + innerPadding)
                    .outerRadius(OR - outerPadding - 0.5 * levelmarkerSW)
                    .startAngle(function(d) {
                        return yscale(d[0]);
                    });
            } else {
                let activeDataCount = _data.reduce((prev, cur) => prev + (cur.disabled ? 0 : 1), 0),
                    interPadding = perToPx(commonPlotOptions.interPadding, OR, 10),
                    band = (OR - IR - (activeDataCount - 1) * interPadding) / activeDataCount,
                    _index = -1;

                levelsInfo.values.forEach((d, i) => {
                    if (!_data[i].disabled) {
                        _index++;
                    }
                    d.innerRadius = _index * (band + interPadding) + IR;
                    d.outerRadius = (_index + 1) * band + _index * interPadding + IR;
                });

                arc.startAngle(commonPlotOptions.startAngle);
                visibility = (d, i) => (_data[i].disabled ? 'hidden' : null);
            }

            var levelmarkerSeries = appendEle(container, 'g', [1], '.', 'class', 'levelmarkerSeries');
            var levelmarker = appendEle(levelmarkerSeries, PATH_ELEMENT, levelMarkerConf.enabled ? levelsInfo.values : []);
            levelmarker
                .attr('class', function(d, i) {
                    return `ZClevelmarker_${i}`;
                })
                .attr('transform', `translate(${plotarea_approx.x},${plotarea_approx.y})`)
                .styles({
                    stroke: function(d, i) {
                        return levelMarkerConf.strokeColor || levelmarkerClr[i];
                    },
                    'stroke-width': levelmarkerSW,
                    'stroke-opacity': levelMarkerConf.strokeOpacity,
                    'fill-opacity': levelMarkerConf.fillOpacity,
                    fill: function(d, i) {
                        var clr = levelmarkerClr[i];
                        var prop = getGradientProp(commonPlotOptions.gradients, [OR, OR], clr);
                        var gradientID = prop
                            ? effects.get(effects.svg_gradients, 'svgGradientFill', prop, { offset: 'pie', chartObj: chartObj })
                            : null;
                        return gradientID ? `url(${defsLocation}#${gradientID})` : clr;
                    },
                    visibility: visibility
                });

            if (isredraw && currentRadiusFactors != previousRadiusFactors) {
                levelmarker
                    .transition()
                    .duration(transitionDuration)
                    .ease(d3_ease_linear)
                    .call(function(element) {
                        arcInnerTween(element, arc, [previousInnerRadius, arc.innerRadius()()], [previousOuterRadius, arc.outerRadius()()]);
                    });
            } else {
                levelmarker.attr('d', arc);
            }

            // add event listeners for dial level marker
            dialUtils.handleEvents(chartObj, levelmarker, levelMarkerConf, 'levelmarker', data);

            if (dataCount > 1 && mode === 'radial') {
                container.selectAll('g.piegroup path').sort((a, b) => a.index - b.index);
            }

            commonRendererProp.commonPlotOptions(commonPlotOptions);
            pie.commonRendererProp(commonRendererProp);
            pie.chartRendered = false;
            container.call(pie);

            if (dataCount > 1 && mode === 'radial') {
                container.selectAll('g.piegroup path').sort((a, b) => d3_descending(valueAccessor(a.data), valueAccessor(b.data)));
            }

            innerRadius = pie.innerRadius();

            // Target marker---
            var targetMarkerSeries = appendEle(container, 'g', [1], '.', 'class', 'targetSeries');
            dialUtils.targetMarker(
                targetMarkerSeries,
                plotarea_approx.actualRadius,
                plotarea_approx.innerRadius,
                chartObj,
                data,
                true,
                undefined,
                { paneid, bound }
            );
            targetMarkerSeries.attr('transform', `translate(${plotarea_approx.x},${plotarea_approx.y})`);
            // ---Target marker

            var needleOptions = commonPlotOptions.needle,
                needleData = _generatedData[0][0].data[0],
                { fillColor: needleFillColor, strokeColor: needleStrokeColor } = needleOptions;

            var dialPath = appendEle(container, PATH_ELEMENT, needleOptions.show ? needleData : [], '.', 'class', 'dial');
            dialPath
                .attr('d', dialUtils.getNeedlePolygon(needleOptions, plotarea_approx, innerRadius))
                .styles({
                    stroke: needleStrokeColor === null ? (d, i) => getColorByIndex(chartObj, i, paneid) : needleStrokeColor || 'black',
                    fill: needleFillColor === null ? (d, i) => getColorByIndex(chartObj, i, paneid) : needleFillColor || NONE,
                    'stroke-width': needleOptions.strokeWidth,
                    visibility: d => (isDisabled(d) ? 'hidden' : null)
                })
                .transition()
                .duration(transitionDuration)
                .ease(easingType)
                .call(dialTween);

            if (labelRenderer) {
                labelQueue.push({
                    renderer: chart,
                    charttype: DIAL_CHARTNAME,
                    chartObj: chartObj,
                    args: [data, 0],
                    rendererIndex,
                    paneid: pick(paneid, 0)
                });
            }
        });
        return chart;
    }

    chart.commonRendererProp = function(_) {
        if (!arguments.length) {
            return commonRendererProp;
        }
        commonRendererProp = _;
        return chart;
    };

    chart.endFunctions = function() {
        previousRadiusFactors = currentRadiusFactors;
        previousOuterRadius = arc.outerRadius()(_data[0]);
        previousInnerRadius = arc.innerRadius()(_data[0]);
        chart.chartRendered = true;
        isredraw = true;
    };

    chart.defaultMouseOutFunc = function() {
        if (chartObj.systemConf.tooltip.enabled) {
            var tooltip = chartObj.eventHandler.tooltipElement;
            tooltip.hide(tooltip.element);
        }
    };

    function createDataForPie(data) {
        if (isFacet(chartObj)) {
            var dt = data[0].data[0];
            data[0].data[0] = [dt[0]];
        }

        return [data];
    }

    function isDisabled(d) {
        return d.disabled;
    }

    function dialTween(element) {
        var cx = plotarea_approx.x,
            cy = plotarea_approx.y,
            baseValue = commonPlotOptions.baseValue !== undefined ? commonPlotOptions.baseValue : 0,
            currentValue,
            previousValue,
            currentAngle,
            previousAngle;

        element.attrTween('transform', function(d, i) {
            if (isredraw) {
                previousAngle = getSvgTransform(this).rotate;
            } else {
                previousValue = hasNegValue ? minMax[1] : minMax[0];
                if (baseValue != null) {
                    previousValue = hasNegValue ? mathMin(baseValue, previousValue) : mathMax(baseValue, previousValue);
                }
                previousAngle = radianToDegree(yscale(previousValue));
            }

            currentValue = mathMin(mathMax(valueAccessor(d), minMax[0]), minMax[1]); //currentValue should be in between min & max
            currentAngle = radianToDegree(yscale(currentValue));

            var interpolateAngle = d3_interpolate(previousAngle, currentAngle);
            return function(t) {
                return `translate(${cx},${cy}) rotate(${interpolateAngle(t)},0,0)`;
            };
        });
    }

    return chart;
};

export var _dialChart = svg_plot_renderer.dial;
