// @zohocharts/zc/src/dial/dialOptions.js
import { defaultTheme as options } from '../core/defaulttheme';
import { mathPI } from '../core/zc-constants';

options.dial = function(userProp) {
    var defaultDialOptions = {
        startAngle: -mathPI / 2,
        endAngle: mathPI / 2,
        innerRadius: '65%',
        maxRadius: 500,
        outerPadding: 20,
        padding: 0,
        interPadding: 10,
        strokeColor: 'rgba(0,0,0,0.1)',
        mode: 'radial',
        animation: {
            type: 'fanAll' //fanAll
        },
        needle: {
            show: true,
            type: 'round', //clock || round || sharp
            strokeColor: 'black',
            fillColor: 'grey',
            width: 15,
            padding: 15
        },
        minmaxlabels: {
            show: true,
            fontSize: userProp.datalabels.fontSize.l5
        },
        datalabels: {
            show: true,
            showAs: 'y', //x,y,percent
            padding: 2
        },
        levelMarker: {
            enabled: true,
            strokeColor: 'rgba(0,0,0,0.1)',
            // "color" : ["#d1d1d1"],
            innerPadding: 0,
            outerPadding: 0,
            colorGamma: [0.3, 0.9]
        },
        targetMarker: {
            outerStrokeWidth: 3,
            symbol: 'line',
            innerFillOpacity: 0,
            innerStrokeOpacity: 0,
            outerFillColor: 'grey',
            outerStrokeColor: '#000'
        },
        gradients: {
            // radial || linear || none
            options: {
                linear: {
                    x2: 2,
                    y2: 2,
                    spreadMethod: 'repeat', // pad || repeat || reflect
                    gradientUnits: 'userSpaceOnUse' //objectBoundingBox || userSpaceOnUse
                },
                radial: {
                    radius: 16.5,
                    spreadMethod: 'reflect' // pad || repeat || reflect
                }
            }
        },
        events: {
            enabled: false
        }
    };
    return defaultDialOptions;
};

export var _dialOptions = options.dial;
