// @zohocharts/zc/src/dial/dial-targetMarker.js
//$Id$
//dial-targetMarker.js
import { dialUtils } from './dial-utils';
import { validateNullVal, defined, pick } from '../utils/zcutils';
import { d0_360, polarToXY, getDY, radianToDegree } from '../polaraxis/polarutils';
import { OUTER_ALIGN, INNER_ALIGN } from '../core/align-constants';
import { appendEle, applyFont, createElement } from '../utils/zcsvgutils';
import { getDashArray, getLineCap } from '../utils/dashstyles';
import { addSVGmarker, getSVGsymbol } from '../utils/svg-symbols';
import { mathMax, mathPI, NONE } from '../core/zc-constants';
import { d3_select } from '../core/d3-constants';
import { getFormattedValue } from '../formatter/formatter';
import { $ZCG } from '../core/zcglobal';
import { isFacet } from '../axis/hierarchy-axis-utils';
import { handleOverlap, getLabelBounds } from '../legend/size-legend/nested/size-nested-ticklabel-handler';

dialUtils.targetMarker = function(container, OR, innerRadius, chartObj, data, handleEvent, labelContainer, options = {}) {
    var systemconf = chartObj.systemConf,
        commonPlotOptions = systemconf.chart.plot.plotoptions.dial,
        dataset = chartObj.dataset,
        index = 0,
        { paneid, bound } = options;

    var yscale = chartObj.axes.y.groups[0].scale,
        minMax = yscale.domain(),
        targetData = data[0].data[0];
    var targetMarkerOptions = commonPlotOptions.targetMarker;
    var align = targetMarkerOptions.align,
        dAlign = targetMarkerOptions.dAlign,
        padding = +(targetMarkerOptions.padding || 0),
        symbol_rotation = /line/.test(targetMarkerOptions.symbol) ? 90 : 0; //For line symbol, the symbol should be rotated by 90 degree

    var default_size = !/(outer|inner)/g.test(align) && /line/.test(targetMarkerOptions.symbol) ? (OR - innerRadius - 2 * padding) / 10 : 1;
    var size = validateNullVal(targetMarkerOptions.size, default_size) * 10;
    var extraOptions = {
        symbolSize: {
            outer: 1,
            inner: 0.5
        },
        color: chartObj.seriesColor[0],
        strokeWidth: default_size
    };
    var dr =
        dAlign === OUTER_ALIGN // dr - shift along radius
            ? size / 2 + padding
            : dAlign === INNER_ALIGN
            ? -(size / 2) - padding
            : 0;
    var r =
        align === OUTER_ALIGN // radius
            ? OR + dr
            : align === INNER_ALIGN
            ? innerRadius + dr
            : (OR + innerRadius) / 2;

    var targetMarker = appendEle(container, 'g', targetMarkerOptions.enabled ? targetData : [], '.', 'class', 'targetmarker'),
        targetMarkerEventHandler = appendEle(
            container,
            'g',
            targetMarkerOptions.enabled && handleEvent ? targetData : [],
            '.',
            'class',
            'targetmarker-eventhandler'
        );

    targetMarker
        .attr('transform', getPosition)
        .styles({
            display: getDisplay,
            'stroke-dasharray': getDashArray(targetMarkerOptions.dashStyle, validateNullVal(targetMarkerOptions.outerStrokeWidth, 1)),
            'stroke-linecap': getLineCap(targetMarkerOptions.lineCap, targetMarkerOptions.dashStyle)
        })
        .each(function() {
            var ele = d3_select(this);
            addSVGmarker(ele, null, targetMarkerOptions, extraOptions, chartObj);
        });

    // add event listeners for dial target marker
    drawEventHandler(targetMarkerEventHandler);
    dialUtils.handleEvents(chartObj, targetMarkerEventHandler, targetMarkerOptions, 'targetmarker', data);

    var targetLabelOptions = targetMarkerOptions.label || {};
    var targetLabelPadding = validateNullVal(targetLabelOptions.padding, 2);
    var targetLabelRadius =
        align === OUTER_ALIGN
            ? r + size / 2 + targetLabelPadding
            : align === INNER_ALIGN
            ? (dAlign === OUTER_ALIGN ? innerRadius : r - size / 2) - targetLabelPadding
            : OR + targetLabelPadding;

    if (!labelContainer) {
        let labelHolder = chartObj.staticPaneSVG.selectAll('g.labelholder'),
            g = createElement('g', '.targetSeries', [1], labelHolder),
            plotarea_approx = chartObj.plotarea_approx;

        if (isFacet(chartObj)) {
            labelContainer = createElement('g', '.target_grid' + paneid, [1], g);
            labelContainer.attr('transform', `translate(${bound.x + plotarea_approx.x || 0},${bound.y + plotarea_approx.y || 0})`);
        } else {
            labelContainer = createElement('g', '.target_0', [1], g);
            labelContainer.attr('transform', `translate(${plotarea_approx.x || 0},${plotarea_approx.y || 0})`);
        }
    }

    var targetLabel = appendEle(
        labelContainer,
        'text',
        targetMarkerOptions.enabled && pick(targetLabelOptions.enabled, true) && !chartObj.renderer[0].firedList.targetSeries
            ? targetData
            : []
    );
    applyFont(targetLabel, [targetLabelOptions, systemconf.chart, systemconf.canvas, $ZCG]);

    targetLabel.each(function(d) {
        var ele = d3_select(this);
        ele.style('display', NONE);

        var targetValue = d[targetMarkerOptions.dataindex];
        if (defined(targetValue)) {
            targetValue = dialUtils.getValueWithinRange(targetValue, minMax);
            if (!(!JSON.parse(commonPlotOptions.minmaxlabels.show) || minMax.indexOf(targetValue) == -1) || isDisabled(d)) {
                return;
            }
            var angle = yscale(targetValue);
            var angle_radian = d0_360(angle);
            var xy = polarToXY(targetLabelRadius, angle);

            var columns = chartObj.userdata.metadata.columns;
            var columnidx = columns.filter(function(d) {
                return d.dataindex === targetMarkerOptions.dataindex;
            });
            var targetColumn = columnidx.length ? columnidx[0] : columns[dataset.getColumnIdx('y', 0)];
            var datatype = targetColumn.datatype;
            var formatConf = targetColumn[datatype];

            var TEXTANCHOR =
                    angle_radian === 0
                        ? 'middle'
                        : align === INNER_ALIGN
                        ? angle_radian < 0 || angle_radian > mathPI
                            ? null
                            : 'end'
                        : angle_radian < 0 || angle_radian > mathPI
                        ? 'end'
                        : null,
                DY =
                    align === INNER_ALIGN
                        ? !(angle_radian > mathPI / 2 && angle_radian < (3 * mathPI) / 2)
                            ? `${-getDY(angle_radian)}em`
                            : null
                        : angle_radian > mathPI / 2 && angle_radian < (3 * mathPI) / 2
                        ? `${getDY(angle_radian)}em`
                        : null,
                TEXT = defined(targetLabelOptions.text)
                    ? dataset.getCustomizedVal(chartObj, targetLabelOptions.text, index, null, null, d, 'targetLabel')
                    : getFormattedValue(formatConf, targetValue, datatype, chartObj, 'targetLabel');

            ele.text(TEXT)
                .style('display', null)
                .attrs({
                    'text-anchor': TEXTANCHOR,
                    dy: DY,
                    transform: `translate(${xy[0]},${xy[1]})`
                });
        }
    });

    let labelNodes = targetLabel
            .filter(function() {
                return d3_select(this).style('display') !== NONE;
            })
            .nodes()
            .reverse(),
        eleBounds = getLabelBounds(labelNodes),
        overlappingHandler = (ele, i) => {
            d3_select(ele).style('display', NONE);
            eleBounds[i] = { x1: 0, y1: 0, x2: 0, y2: 0 };
        };
    handleOverlap(labelNodes, eleBounds, overlappingHandler);

    function getPosition(d) {
        var targetValue = d[targetMarkerOptions.dataindex];
        if (defined(targetValue)) {
            targetValue = dialUtils.getValueWithinRange(targetValue, minMax);
            var angle_radian = yscale(targetValue),
                xy = polarToXY(r, angle_radian),
                angle_degree = radianToDegree(angle_radian) + symbol_rotation;
            return `translate(${xy[0]},${xy[1]}) rotate(${angle_degree},0,0)`;
        }
    }

    function getDisplay(d) {
        var targetValue = d[targetMarkerOptions.dataindex];
        return !defined(targetValue) || isDisabled(d) ? NONE : null;
    }

    function isDisabled(d) {
        return d.disabled;
    }

    function drawEventHandler(element) {
        var triggerOffset = (targetMarkerOptions.events && targetMarkerOptions.events.triggerOffset) || 0;
        extraOptions.strokeWidth = (size + triggerOffset * 2) / 10;

        element
            .attrs({
                transform: getPosition,
                id: (d, i) => `ZCtargetmarker_eventhandler_${i}`
            })
            .styles({
                display: getDisplay,
                opacity: 0
            })
            .each(function() {
                var ele = d3_select(this);
                addSVGmarker(ele, null, targetMarkerOptions, extraOptions, chartObj);
                if (targetMarkerOptions.symbol === 'line') {
                    var w = size >= OR - innerRadius ? size : mathMax(size + triggerOffset, OR - innerRadius),
                        h = triggerOffset || 1; // If triggerOffset = 0, use height as 1 (As stroke properties are not considered if height is 0)
                    ele.selectAll('*').attr('d', getSVGsymbol('rect')(w, h));
                }
            });
    }
};

export var _dialTargetMarker = dialUtils.targetMarker;
