// @zohocharts/zc/src/dial/dial-events.js
//$Id$
//dial-events.js
import { dialUtils } from './dial-utils';
import { validateNullVal, invokeFunction } from '../utils/zcutils';
import { getTooltipPosition, getEventObject } from '../events/events-utils';
import { POINTER, NULL } from '../core/zc-constants';
import { attachevents, getMousePosition } from '../events/eventActions';
import { getFormattedValue } from '../formatter/formatter';
import { docTapFunction } from '../events/document-events';
import { d3_event } from '../core/d3-constants';

dialUtils.handleEvents = function(chartObj, element, options, type, data) {
    var currentPoint = data[0].data[0][0],
        isLevelMarker = type === 'levelmarker',
        levelMarkerInstance = chartObj.renderer[0].levelMarker,
        levelMarkerLength = levelMarkerInstance && levelMarkerInstance.values.length,
        commonPlotOptions = chartObj.systemConf.chart.plot.plotoptions.dial,
        tooltipEnabled = chartObj.systemConf.tooltip.enabled,
        events_dial = options.events || {},
        metaInfo,
        formatConf,
        datatype;

    var isEventEnabled = validateNullVal(events_dial.enabled, commonPlotOptions.events.enabled);
    var cursor = isEventEnabled ? events_dial.cursor || POINTER : null;

    element.style('cursor', cursor);

    if (isEventEnabled) {
        var dataset = chartObj.dataset,
            columns = chartObj.userdata.metadata.columns,
            totalData = data[0].data[0].length,
            columnidx = columns.filter(function(d) {
                return d.dataindex === options.dataindex;
            }),
            targetColumn = columnidx.length ? columnidx[0] : columns[dataset.getColumnIdx('y', 0)],
            defaultMouseOverFunc = function(d, i) {
                var val = isLevelMarker ? d[1] : d[options.dataindex],
                    userCallback = createUserObject(d),
                    position = getTooltipPosition(
                        d3_event(),
                        {
                            x: 0,
                            y: 0
                        },
                        0,
                        0,
                        chartObj,
                        {
                            seriesIndex: 0,
                            subseriesIndex: 0,
                            itemIndex: i
                        }
                    );

                if (d._color) {
                    position.helpers.color = d._color;
                }

                if (tooltipEnabled) {
                    var tooltipContent =
                        invokeFunction(events_dial.tooltipContent, userCallback, chartObj) ||
                        (isLevelMarker ? 'Range: ' : 'Target: ') + userCallback.formattedValue;
                    var tooltip = chartObj.eventHandler.tooltipElement;
                    tooltip.show(chartObj, getTooltipPositionForNonAxisCharts(), tooltipContent, false, position.helpers);
                }
            },
            defaultMouseOutFunc = function() {
                if (tooltipEnabled) {
                    var tooltip = chartObj.eventHandler.tooltipElement;
                    tooltip.hide(tooltip.element);
                }
            };
        datatype = targetColumn.datatype;
        formatConf = targetColumn[datatype];

        var mousemove = function(e, d) {
                var previousChartObj = NULL; //#ZC1228
                var tooltip = chartObj.eventHandler.tooltipElement;
                if (tooltip && tooltip.timeout) {
                    clearTimeout(tooltip.timeout); // To avoid flickering on  mouse move, clearing the timeout that makes the tooltip's display prop to none
                }

                if (chartObj.stopPropogate !== false) {
                    d3_event(e).stopPropagation();
                }

                var eventType = chartObj.zcEventType || 'mousemove';
                handleUserCallback(currentPoint, eventType, d3_event(e), d);
                defaultMouseOverFunc(d, d.index);
            },
            mouseout = function(e, d) {
                var eventType = chartObj.zcEventType || 'mouseout';
                handleUserCallback(currentPoint, eventType, d3_event(e), d);
                defaultMouseOutFunc();
            },
            click = function(e, d) {
                const i = d.index;
                var isTouchEvent = d3_event(e).touches,
                    eventType = chartObj.zcEventType || 'click',
                    itemPos = totalData + i + 1 + (type === 'targetmarker' ? levelMarkerLength : 0);

                handleUserCallback(currentPoint, eventType, e, d);
                if (isTouchEvent && eventType !== 'taphold') {
                    docTapFunction(chartObj);
                    // var clickedTime = new Date().getTime(),
                    //     clickedElement = this,
                    //     clickArguments = [chartObj.renderer[0], d, 0, itemPos],
                    //     previousChartObj = chartObj;

                    var tapOldInfo = chartObj.eventHandler.tapOldInfo;
                    if (!tapOldInfo || tapOldInfo.itemPos != itemPos) {
                        defaultMouseOverFunc(d, i);
                        chartObj.eventHandler.tapOldInfo = {
                            seriesPos: 0,
                            subseriesindex: 0,
                            itemPos: itemPos,
                            type: type
                        };
                    } else {
                        defaultMouseOutFunc();
                        chartObj.eventHandler.tapOldInfo = null;
                    }
                }
            },
            dblclick = function(e, d) {
                var eventType = chartObj.zcEventType || 'doubleclick';
                handleUserCallback(currentPoint, eventType, d3_event(e), d);
            },
            eventAttacherOptions = {
                target: 'plot',
                touchleave: true,
                chartObj: chartObj
            };
        var mouseHandlers = {
                mousemove: mousemove,
                mouseout: mouseout,
                click: click,
                dblclick: dblclick
            },
            touchHandlers = {
                singletouchmove: mousemove,
                touchend: mouseout,
                tap: click,
                dbltap: dblclick
            };
        attachevents(element, mouseHandlers, touchHandlers, eventAttacherOptions);
    }

    function getTooltipPositionForNonAxisCharts() {
        var mouse = getMousePosition(chartObj.svg.node(), d3_event());
        var relmousePt = {
            x: mouse[0],
            y: mouse[1]
        };
        var tooltipPos = {
            left: relmousePt.x,
            top: relmousePt.y
        };
        var evt = getEventObject(d3_event());
        var dLeft = relmousePt.x - evt.clientX,
            dTop = relmousePt.y - evt.clientY;
        tooltipPos.left -= dLeft;
        tooltipPos.top -= dTop;
        tooltipPos.dLeft = dLeft;
        tooltipPos.dTop = dTop;
        return tooltipPos;
    }

    function handleUserCallback(data, type, event, d) {
        return invokeFunction(events_dial[type], event, createUserObject(d), chartObj);
    }

    function createUserObject(d) {
        var val = isLevelMarker ? d[1] : d[options.dataindex];
        return {
            value: val,
            formattedValue: getFormattedValue(formatConf, val, datatype, chartObj),
            actualData: currentPoint,
            data: (metaInfo = metaInfo || chartObj.dataset.getMetaInfoObj(chartObj, currentPoint, 0))
        };
    }
};

export var _dialEvents = dialUtils.handleEvents;
