// @zohocharts/zc/src/sankey/d3-sankey-polyfill.js
import { Registry } from '../core/zc-registry';
import { d3_linkHorizontal, d3_map, d3_max, d3_min, d3_sum, d3_mean } from '../core/d3-constants';
import { circularSankeyGenerator } from './sankey-circular-polyfill';
import { getCircularLinkMargin } from './sankey-circular-tools';
import { getDepth, getPropVal } from './../utils/zcutils';

function targetDepth(d) {
    return d.target.depth;
}

function left(node) {
    return node.depth;
}

function right(node, n) {
    return n - 1 - node.height;
}

function justify(node, n) {
    return node.sourceLinks.length ? node.depth : n - 1;
}

function center(node) {
    return node.targetLinks.length ? node.depth : node.sourceLinks.length ? d3_min(node.sourceLinks, targetDepth) - 1 : 0;
}

/**
 * This function makes positioning of nodes based on level value
 * @param {Array} node - array of sankey nodes
 * @param {Number} n - maximum depth of sankey node
 * @returns {Number}
 */
function level(node, n) {
    return node.level ? getDepth(node.level) - 1 : center(node);
}

function constant(x) {
    return function() {
        return x;
    };
}

function ascendingSourceBreadth(a, b) {
    return ascendingBreadth(a.source, b.source) || a.index - b.index;
}

function ascendingTargetBreadth(a, b) {
    return ascendingBreadth(a.target, b.target) || a.index - b.index;
}

function ascendingBreadth(a, b) {
    if (a.partOfCycle === b.partOfCycle) {
        return a.y0 - b.y0;
    } else {
        if (a.circularLinkType === 'top' || b.circularLinkType === 'bottom') {
            return -1;
        } else {
            return 1;
        }
    }
}

function value(d) {
    return d.value;
}

function nodeCenter(node) {
    return (node.y0 + node.y1) / 2;
}

function linkSourceCenter(link) {
    return nodeCenter(link.source);
}

function linkTargetCenter(link) {
    return nodeCenter(link.target);
}

function defaultId(d) {
    return d.index;
}

function defaultNodes(graph) {
    return graph.nodes;
}

function defaultLinks(graph) {
    return graph.links;
}

function find(nodeById, id) {
    let node = nodeById.get(id);

    if (!node) {
        throw new Error(`missing: ${id}`);
    }

    return node;
}
let verticalMargin = 25;
let baseRadius = 10;

var Sankey = function() {
    let x0 = 0,
        y0 = 0,
        x1 = 1,
        y1 = 1, // extent
        dx = 24, // nodeWidth
        py, // nodePadding
        dy = 8,
        scale = 1,
        id = defaultId,
        align = justify,
        nodes = defaultNodes,
        links = defaultLinks,
        link_padding = 0,
        iterations = 1;
    let sort;
    let linkSort;
    let chartObj;

    let padding = Infinity;
    let paddingRatio = 0.1,
        height,
        margin,
        sortNodes = null;

    function sankey() {
        var graph = {
            nodes: nodes.apply(null, arguments),
            links: links.apply(null, arguments),
            matrix: {},
            x0: x0,
            y0: y0,
            x1: x1,
            y1: y1
        };
        var sankeyExtent = {
            x0: x0,
            y0: y0,
            x1: x1,
            y1: y1
        };
        computeNodeLinks(graph);
        identifyCircles(graph);
        selectCircularLinkTypes(graph);
        computeNodeValues(graph);

        computeNodeDepths(graph);
        adjustSankeySize(graph, sankeyExtent);
        computeNodeBreadths(graph, iterations);
        computeLinkBreadths(graph, sankeyExtent);

        if (graph.circular) {
            circularSankeyGenerator(graph, sankeyExtent, link_padding);
        }

        return graph;
    }

    sankey.update = function(graph) {
        computeLinkBreadths(graph);
        return graph;
    };

    sankey.nodeId = function(_) {
        return arguments.length ? ((id = typeof _ === 'function' ? _ : constant(_)), sankey) : id;
    };

    sankey.nodeAlign = function(_) {
        return arguments.length ? ((align = typeof _ === 'function' ? _ : constant(_)), sankey) : align;
    };

    sankey.nodeSort = function(_) {
        return arguments.length ? ((sort = _), sankey) : sort;
    };

    sankey.nodeWidth = function(_) {
        return arguments.length ? ((dx = +_), sankey) : dx;
    };

    sankey.nodePadding = function(_) {
        return arguments.length ? ((dy = py = +_), sankey) : dy;
    };
    sankey.linkPadding = function(_) {
        return arguments.length ? ((link_padding = +_), sankey) : link_padding;
    };

    sankey.scale = function(_) {
        return arguments.length ? ((scale = +_), sankey) : scale;
    };

    sankey.nodes = function(_) {
        return arguments.length ? ((nodes = typeof _ === 'function' ? _ : constant(_)), sankey) : nodes;
    };

    sankey.links = function(_) {
        return arguments.length ? ((links = typeof _ === 'function' ? _ : constant(_)), sankey) : links;
    };

    sankey.linkSort = function(_) {
        return arguments.length ? ((linkSort = _), sankey) : linkSort;
    };

    sankey.size = function(_) {
        return arguments.length ? ((x0 = y0 = 0), (x1 = +_[0]), (y1 = +_[1]), sankey) : [x1 - x0, y1 - y0];
    };

    sankey.extent = function(_) {
        return arguments.length ? ((x0 = +_[0][0]), (x1 = +_[1][0]), (y0 = +_[0][1]), (y1 = +_[1][1]), sankey) : [[x0, y0], [x1, y1]];
    };

    sankey.iterations = function(_) {
        return arguments.length ? ((iterations = +_), sankey) : iterations;
    };

    sankey.nodePaddingRatio = function(_) {
        return arguments.length ? ((paddingRatio = +_), sankey) : paddingRatio;
    };

    sankey.height = function(_) {
        return arguments.length ? ((height = +_), sankey) : height;
    };
    sankey.margin = function(_) {
        return arguments.length ? ((margin = _), sankey) : margin;
    };
    sankey.chart = function (_) {
        return arguments.length ? (chartObj = _, sankey) : chartObj;
    }

    // Populate the sourceLinks and targetLinks for each node.
    // Also, if the source and target are not objects, assume they are indices.
    function computeNodeLinks(graph) {
        graph.nodes.forEach(function(node, i) {
            node.index = i;
            node.sourceLinks = [];
            node.targetLinks = [];
        });
        var nodeById = d3_map(graph.nodes, id);
        graph.links.forEach(function(link, i) {
            link.index = i;
            var source = link.source;
            var target = link.target;
            if (typeof source !== 'object') {
                source = link.source = find(nodeById, source);
            }
            if (typeof target !== 'object' && !link.exit) {
                target = link.target = find(nodeById, target);
            }
            source.sourceLinks.push(link);
            if (!link.exit) {
                source.hasExitLink = true;
                target.targetLinks.push(link);
            }
        });

        graph.nodes.forEach(function(node, i) {
            if (node.hasExitLink) {
                node.sourceLinks.sort(function(a, b) {
                    return a.exit ? 100000 - b.index : b.exit ? a.index - 100000 : a.index - b.index;
                });
            }
        });
    }

    // Compute the value (size) and cycleness of each node by summing the associated links.
    function computeNodeValues(graph) {
        graph.nodes.forEach(function(node) {
            node.partOfCycle = false;
            node.value = Math.max(d3_sum(node.sourceLinks, value), d3_sum(node.targetLinks, value));
            node.sourceLinks.forEach(function(link) {
                if (link.circular) {
                    node.partOfCycle = true;
                    node.circularLinkType = link.circularLinkType;
                }
            });
            node.targetLinks.forEach(function(link) {
                if (link.circular) {
                    node.partOfCycle = true;
                    node.circularLinkType = link.circularLinkType;
                }
            });
        });
    }

    // Iteratively assign the depth (x-position) for each node.
    // Nodes are assigned the maximum depth of incoming neighbors plus one;
    // nodes with no incoming links are assigned depth zero, while
    // nodes with no outgoing links are assigned the maximum depth.
    function computeNodeDepths(graph) {
        var nodes, next, x;

        for (nodes = graph.nodes, next = [], x = 0; nodes.length; ++x, nodes = next, next = []) {
            for (let i in nodes) {
                if (nodes.hasOwnProperty(i)) {
                    let node = nodes[i];
                    node.depth = x;
                    for (let j in node.sourceLinks) {
                        if (node.sourceLinks.hasOwnProperty(j)) {
                            let link = node.sourceLinks[j];
                            if (next.indexOf(link.target) < 0 && !link.circular && !link.exit) {
                                next.push(link.target);
                            }
                        }
                    }
                }
            }
        }

        for (nodes = graph.nodes, next = [], x = 0; nodes.length; ++x, nodes = next, next = []) {
            for (let i in nodes) {
                if (nodes.hasOwnProperty(i)) {
                    let node = nodes[i];
                    node.height = x;
                    for (let j in node.targetLinks) {
                        if (node.targetLinks.hasOwnProperty(j)) {
                            let link = node.targetLinks[j];
                            if (next.indexOf(link.source) < 0 && !link.circular) {
                                next.push(link.source);
                            }
                        }
                    }
                }
            }
        }

        var kx = (x1 - x0 - dx) / (x - 1);
        graph.nodes.forEach(function(node) {
            node.x1 = (node.x0 = x0 + Math.max(0, Math.min(x - 1, Math.floor(align.call(null, node, x)))) * kx) + dx;
        });
        // assign column numbers, and get max value
        graph.nodes.forEach(function(node) {
            node.column = sortNodes !== null ? node[sortNodes] : Math.floor(align.call(null, node, x));
        });
    }
    function computeNodeLayers(graph) {
        let nodes = graph.nodes,
            x = d3_max(nodes, d => d.depth) + 1,
            kx = (x1 - x0 - dx) / (x - 1);

        let columns = new Array(x);
        nodes.forEach(node => {
            let i = Math.max(0, Math.min(x - 1, Math.floor(align.call(null, node, x))));
            node.layer = i;
            node.x0 = x0 + i * kx;
            node.x1 = node.x0 + dx;

            if (columns[i]) {
                columns[i].push(node);
            } else {
                columns[i] = [node];
            }
        });

        if (sort) {
            //get sort callback function from string value
            if(typeof sort === 'string') {
                var temp = getPropVal(window, sort);
                sort = typeof temp === 'function' ? temp : undefined;
            }
            columns.forEach(column => {
                column.sort(function () { return sort.apply(this, [...arguments, chartObj]); });
            });
        }

        return columns;
    }
    function computeNodeBreadths(graph) {
        let columns = computeNodeLayers(graph),
            maxRows = d3_max(columns, c => c.length);
        graph.columns = columns.length;
        graph.rows = maxRows;
        graph.matrix.nodes = columns;
        py = Math.min(dy, (y1 - y0) / (maxRows - 1));

        initializeNodeBreadth();

        for (let i = 0; i < iterations; ++i) {
            const alpha = Math.pow(0.99, i);
            const beta = Math.max(1 - alpha, (i + 1) / iterations);
            relaxRightToLeft(columns, alpha, beta);
            relaxLeftToRight(columns, alpha, beta);
        }
        // Reposition each node based on its incoming (target) links.
        function _loop1(column, i, alpha, beta){
            if (column.hasOwnProperty(i)) {
                let target = column[i];
                let y = 0;
                let w = 0;
                target.targetLinks.forEach(targetLink => {
                    let { source, value } = targetLink,
                        v = value * (target.layer - source.layer);
                    y += targetTop(source, target) * v;
                    w += v;
                });
                if (!(w > 0)) {
                    let __x = 0;
                }else{
                    let dy = (y / w - target.y0) * alpha;
                    target.y0 += dy;
                    target.y1 += dy;
                    reorderNodeLinks(target);
                }
                
            }
        }
        function relaxLeftToRight(columns, alpha, beta) {
            for (let i = 1, n = columns.length; i < n; ++i) {
                const column = columns[i];
                for (let i in column) {
                    _loop1(column, i, alpha, beta);
                }
                if (sort === undefined) {
                    column.sort(ascendingBreadth);
                }
                resolveCollisions(column, beta, graph.circular);
            }
        }

        
        function _loop2(column, i, alpha, beta){
            if (column.hasOwnProperty(i)) {
                let source = column[i];
                let y = 0;
                let w = 0;
                source.sourceLinks.forEach(sourceLink => {
                    let { target, value } = sourceLink,
                        v = value * (target.layer - source.layer);
                    y += sourceTop(source, target) * v;
                    w += v;
                });
                if (!(w > 0)) {
                    let __x = 0;
                }else{
                    let dy = (y / w - source.y0) * alpha;
                    source.y0 += dy;
                    source.y1 += dy;
                    reorderNodeLinks(source);
                }
                
            }
        }
        // Reposition each node based on its outgoing (source) links.
        function relaxRightToLeft(columns, alpha, beta) {
            for (let n = columns.length, i = n - 2; i >= 0; --i) {
                const column = columns[i];
                for (let i in column) {
                    _loop2(column, i, alpha, beta);
                }
                if (sort === undefined) {
                    column.sort(ascendingBreadth);
                }
                resolveCollisions(column, beta);
            }
        }
        

        function resolveCollisions(nodes, alpha) {
            const i = nodes.length >> 1;
            const subject = nodes[i];
            resolveCollisionsBottomToTop(nodes, subject.y0 - py, i - 1, alpha);
            resolveCollisionsTopToBottom(nodes, subject.y1 + py, i + 1, alpha);
            resolveCollisionsBottomToTop(nodes, y1, nodes.length - 1, alpha);
            resolveCollisionsTopToBottom(nodes, y0, 0, alpha);
        }

        // Push any overlapping nodes down.
        function resolveCollisionsTopToBottom(nodes, y, i, alpha) {
            for (; i < nodes.length; ++i) {
                const node = nodes[i];
                const dy = (y - node.y0) * alpha;
                if (dy > 1e-6) {
                    (node.y0 += dy), (node.y1 += dy);
                }
                y = node.y1 + py;
            }
        }

        // Push any overlapping nodes up.
        function resolveCollisionsBottomToTop(nodes, y, i, alpha) {
            for (; i >= 0; --i) {
                const node = nodes[i];
                const dy = (node.y1 - y) * alpha;
                if (dy > 1e-6) {
                    (node.y0 -= dy), (node.y1 -= dy);
                }
                y = node.y0 - py;
            }
        }

        function reorderNodeLinks({ sourceLinks, targetLinks }) {
            if (linkSort === undefined) {
                if (targetLinks) {
                    targetLinks.forEach(targetLink => {
                        const {
                            source: { sourceLinks }
                        } = targetLink;
                        sourceLinks.sort(ascendingTargetBreadth);
                    });
                }
                sourceLinks.forEach(sourceLink => {
                    const {
                        target: { targetLinks }
                    } = sourceLink;
                    if (targetLinks) {
                        targetLinks.sort(ascendingSourceBreadth);
                    }
                });
            }
        }
        function reorderLinks(nodes) {
            if (linkSort === undefined) {
                nodes.forEach(node => {
                    const { sourceLinks, targetLinks } = node;
                    sourceLinks.sort(ascendingTargetBreadth);
                    if (targetLinks) {
                        targetLinks.sort(ascendingSourceBreadth);
                    }
                });
            }
        }
        // Returns the target.y0 that would produce an ideal link from source to target.
        function targetTop(source, target) {
            let y = source.y0 - ((source.sourceLinks.length - 1) * py) / 2;
            for (let i in source.sourceLinks) {
                if (source.sourceLinks.hasOwnProperty(i)) {
                    let { target: node, width } = source.sourceLinks[i];
                    if (node === target) {
                        break;
                    }
                    y += width + py;
                }
            }
            for (let i in target.targetLinks) {
                if (target.targetLinks.hasOwnProperty(i)) {
                    const { source: node, width } = target.targetLinks[i];
                    if (node === source) {
                        break;
                    }
                    y -= width;
                }
            }
            return y;
        }

        // Returns the source.y0 that would produce an ideal link from source to target.
        function sourceTop(source, target) {
            if (target != 'null') {
                let y = target.y0 - ((target.targetLinks.length - 1) * py) / 2;
                for (let i in target.targetLinks) {
                    if (target.targetLinks.hasOwnProperty(i)) {
                        const { source: node, width } = target.targetLinks[i];
                        if (node === source) {
                            break;
                        }
                        y += width + py;
                    }
                }
                for (let i in source.sourceLinks) {
                    if (source.sourceLinks.hasOwnProperty(i)) {
                        const { target: node, width } = source.sourceLinks[i];
                        if (node === target) {
                            break;
                        }
                        y -= width;
                    }
                }
                return y;
            }
        }
        function initializeNodeBreadth() {
            // columns.forEach(function(nodes) {
            //     let thisPadding = (y1 * paddingRatio) / (nodes.length + 1);
            //     padding = thisPadding < padding ? thisPadding : padding;
            // });
            //
            // py = padding;

            var ky = d3_min(columns, c => (y1 - y0 - (c.length - 1) * py) / d3_sum(c, value));
            var ky1 = d3_min(columns, c => (y1 - y0 - (c.length - 1) * py) / d3_sum(c, value));

            ky = graph.ky || ky;
            // ky = ky * scale;

            columns.forEach(nodes => {
                let y = y0;
                nodes.forEach(node => {
                    node.y0 = y;
                    node.y1 = y + node.value * ky;
                    y = node.y1 + py;
                    node.sourceLinks.forEach(link => {
                        link.width = link.value * ky;
                    });
                });
                y = (y1 - y + py) / (nodes.length + 1);
                for (let i = 0; i < nodes.length; ++i) {
                    const node = nodes[i];
                    if (node.partOfCycle) {
                        if (node.circularLinkType == 'top') {
                            node.y0 = y0 + i;
                            node.y1 = node.y0 + node.value * ky;
                        } else {
                            node.y0 = y1 - node.value * ky - i;
                            node.y1 = node.y0 + node.value * ky;
                        }
                    } else {
                        node.y0 += y * (i + 1);
                        node.y1 += y * (i + 1);
                    }
                }
                reorderLinks(nodes);
            });
        }

        function relaxLeftAndRight(alpha) {
            let columnsLength = columns.length;
            // console.log("cols: " + columnsLength);

            columns.forEach(function(nodes, i) {
                let n = nodes.length;
                let depth = nodes[0].depth;

                // console.log('~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~')

                // console.log(depth + ': ' + n)

                nodes.forEach(function(node) {
                    // check the node is not an orphan
                    if (node.sourceLinks.length || node.targetLinks.length) {
                        if (!node.partOfCycle && depth == 0 && n == 1) {
                            /*
                            if (node.partOfCycle)
                            // console.log('do nothing')
                             */
                            let nodeHeight = node.y1 - node.y0;

                            node.y0 = y1 / 2 - nodeHeight / 2;
                            node.y1 = y1 / 2 + nodeHeight / 2;
                        } else if (depth == columnsLength - 1 && n == 1) {
                            let nodeHeight = node.y1 - node.y0;

                            node.y0 = y1 / 2 - nodeHeight / 2;
                            node.y1 = y1 / 2 + nodeHeight / 2;
                        } else {
                            let avg = 0;

                            let avgTargetY = d3_mean(node.sourceLinks, linkTargetCenter);
                            let avgSourceY = d3_mean(node.targetLinks, linkSourceCenter);

                            if (avgTargetY && avgSourceY) {
                                avg = (avgTargetY + avgSourceY) / 2;
                            } else {
                                avg = avgTargetY || avgSourceY;
                            }

                            let dy = (avg - nodeCenter(node)) * alpha;

                            // positive if it node needs to move down
                            // let dy = (nodeCenter(node.sourceLinks[0].target) - nodeCenter(node.sourceLinks[0].source) / 2) * alpha;

                            node.y0 += dy;
                            node.y1 += dy;

                            // console.log('after: ' + node.y0 + ' ' + node.y1)
                        }
                    }
                });
            });
        }
        function resolveCollisions(nodes, alpha, circular) {
            if (circular) {
                return resolveCollisions1();
            }
            let i = nodes.length >> 1;
            let subject = nodes[i];
            resolveCollisionsBottomToTop(nodes, subject.y0 - py, i - 1, alpha);
            resolveCollisionsTopToBottom(nodes, subject.y1 + py, i + 1, alpha);
            resolveCollisionsBottomToTop(nodes, y1, nodes.length - 1, alpha);
            resolveCollisionsTopToBottom(nodes, y0, 0, alpha);
        } // Push any overlapping nodes down.
        function resolveCollisions1() {
            // for (; i >= 0; --i) {
            //     const node = nodes[i];
            //     const dy = (node.y1 - y) * alpha;
            //     if (dy > 1e-6) {
            //         (node.y0 -= dy), (node.y1 -= dy);
            //     }
            //     y = node.y0 - py;
            // }
            columns.forEach(function(nodes) {
                var node,
                    dy,
                    y = y0,
                    n = nodes.length,
                    i;

                // Push any overlapping nodes down.
                nodes.sort(ascendingBreadth);

                for (i = 0; i < n; ++i) {
                    node = nodes[i];
                    dy = y - node.y0;

                    if (dy > 0) {
                        node.y0 += dy;
                        node.y1 += dy;
                    }
                    y = node.y1 + py;
                }

                // If the bottommost node goes outside the bounds, push it back up.
                dy = y - py - y1;
                if (dy > 0) {
                    (y = node.y0 -= dy), (node.y1 -= dy);

                    // Push any overlapping nodes back up.
                    for (i = n - 2; i >= 0; --i) {
                        node = nodes[i];
                        dy = node.y1 + py - y;
                        if (dy > 0) {
                            (node.y0 -= dy), (node.y1 -= dy);
                        }
                        y = node.y0;
                    }
                }
            });
        }
    }

    function computeLinkBreadths(graph, sankeyExtent) {
        graph.nodes.forEach(function(node) {
            node.sourceLinks.sort(ascendingTargetBreadth);
            node.targetLinks.sort(ascendingSourceBreadth);
        });
        graph.nodes.forEach(function(node) {
            var y0 = node.y0;
            var y1 = y0;

            // start from the bottom of the node for cycle links
            var y0cycle = node.y1;
            var y1cycle = y0cycle;

            node.sourceLinks.forEach(function(link) {
                if (link.circular) {
                    link.y0 = y0cycle - link.width / 2;
                    y0cycle = y0cycle - link.width;
                } else {
                    link.y0 = y0 + link.width / 2;
                    y0 += link.width;
                }
                //Precalculate exit node at end node details
                if (link.exit && !graph.exitAtEndNode && sankeyExtent) {
                    let s = link.source;
                    if (s.x1 == sankeyExtent.x1) {
                        graph.exitAtEndNode = link.width;
                    }
                }
            });
            node.targetLinks.forEach(function(link) {
                if (link.circular) {
                    link.y1 = y1cycle - link.width / 2;
                    y1cycle = y1cycle - link.width;
                } else {
                    link.y1 = y1 + link.width / 2;
                    y1 += link.width;
                }
            });
        });
    }
    function adjustSankeySize(inputGraph, sankeyExtent) {
        // return
        let graph = inputGraph;
        if (graph.circular) {
            let columns = computeNodeLayers(graph);

            graph.py = py;
            var ky = d3_min(columns, c => (sankeyExtent.y1 - sankeyExtent.y0 - (c.length - 1) * py) / d3_sum(c, value));

            // graph.links.forEach(function(link) {
            //     link.width = link.value * ky;
            // });

            var maxColumn = d3_max(graph.nodes, function(node) {
                return node.column;
            });
            //determine how much to scale down the chart, based on circular links
            /*
            PRECALCULATE CIRCULAR LINK MARGINS
             */
            var currentWidth = graph.x1 - graph.x0;
            var currentHeight = graph.y1 - graph.y0;

            var linkmargin = getCircularLinkMargin(graph, ky, null, maxColumn);

            /*
                RECALCULATE Y0, Y1
                 */
            var newWidth = currentWidth + linkmargin.right + linkmargin.left;
            var newHeight = currentHeight + linkmargin.top + linkmargin.bottom;

            var scaleX = currentWidth / newWidth;
            var scaleY = currentHeight / newHeight;

            if (linkmargin.top > currentHeight || linkmargin.top > graph.y1 * scaleY) {
                //if margin values exceed the actual height of sanky, we will calculate proportions between margin top and margin bottom.
                var expectedHei = currentHeight * scaleY;
                var remaining = currentHeight - expectedHei;
                var totalYMargin = linkmargin.top + linkmargin.bottom,
                    topRatio = linkmargin.top / totalYMargin,
                    bottomRatio = linkmargin.bottom / totalYMargin,
                    newTopMargin = remaining * topRatio,
                    newBottomMargin = remaining * bottomRatio;
                y0 = graph.y0 = newTopMargin;
                y1 = graph.y1 = graph.y1 - newBottomMargin;
            } else {
                y0 = graph.y0 = graph.y0 * scaleY + linkmargin.top;
                y1 = graph.y1 = linkmargin.bottom == 0 ? graph.y1 : graph.y1 * scaleY;

                x0 = graph.x0 = graph.x0 * scaleX + linkmargin.left;
                x1 = graph.x1 = linkmargin.right == 0 ? graph.x1 : graph.x1 * scaleX;
            }
            /*
            re-calculate widths BASED ON NEW KY

             */
            // ky = ky * scaleY;
            ky = d3_min(columns, c => (y1 - y0 - (c.length - 1) * py) / d3_sum(c, value));
            // graph.ky = ky;
            var linkmargin = getCircularLinkMargin(graph, ky, null, maxColumn);

            /*
            ADJUSTMENT BASED ON ORIGINAL HEIGHT
            find final estimated sankey height
             */

            var existingY1 = graph.y1;
            var existingY0 = graph.y0;
            // y0 = graph.y0 = sankeyExtent.y0 + linkmargin.top;
            var newHeight = (y1 = graph.y1 = sankeyExtent.y1 - linkmargin.bottom);

            // var y_ratio = (y1 - y0) / (existingY1 - existingY0);
            var y_ratio = newHeight / existingY1;

            // currentHeight = existingY1 - existingY0;
            // scaleY = currentHeight / (currentHeight + linkmargin.top + linkmargin.bottom);

            // y0 = graph.y0 = sankeyExtent.y0 * scaleY + linkmargin.top;
            // y0 = graph.y0 = existingY0 * scaleY + linkmargin.top+existingY0 ;
            // y1 = graph.y1 = (existingY1 - existingY1 * scaleY);

            // var y_ratio = (sankeyExtent.y1 - linkmargin.bottom - sankeyExtent.y0 + linkmargin.top) / (existingY1 - existingY0);
            // var y_ratio = newHeight / existingY1;

            ky = d3_min(columns, c => (y1 - y0 - (c.length - 1) * py) / d3_sum(c, value));
            // graph.ky = ky;

            /*
            Find actual deviated margins from original margin calculated for outside label
            Already sankeyextent.x0 and x1 values loaded with datalabel margin
            Now we need to adjust it with circular link margin
             */
            var linkmargin = getCircularLinkMargin(graph, null, y_ratio, maxColumn);
            if (margin && margin.right) {
                //if outside label margin is >  circular link margin, then no need of correction, else find differenece between the both
                linkmargin.right = margin.right > linkmargin.right ? 0 : linkmargin.right - margin.right;
            }
            if (margin && margin.left) {
                linkmargin.left = margin.left > linkmargin.left ? 0 : linkmargin.left - margin.left;
            }

            currentWidth = sankeyExtent.x1 - sankeyExtent.x0;
            scaleX = currentWidth / (currentWidth + linkmargin.right + linkmargin.left);
            // x0 = graph.x0 = sankeyExtent.x0 * scaleX + linkmargin.left;
            // x1 = graph.x1 = linkmargin.right == 0 ? sankeyExtent.x1 : sankeyExtent.x1 * scaleX;

            x0 = graph.x0 = sankeyExtent.x0 + linkmargin.left;
            x1 = graph.x1 = sankeyExtent.x1 - linkmargin.right;

            // y0 = graph.y0 = sankeyExtent.y0 + linkmargin.top;
            // y1 = graph.y1 = sankeyExtent.y1 - linkmargin.bottom;
            ky = d3_min(columns, c => (y1 - y0 - (c.length - 1) * py) / d3_sum(c, value));
            // graph.ky = ky;
            graph.nodes.forEach(function(node) {
                node.x0 = graph.x0 + node.column * ((graph.x1 - graph.x0 - dx) / maxColumn);
                node.x1 = node.x0 + dx;
            });
        }

        function getCircularLinkMargin1(graph, ky, ratio, maxColumn) {
            var totalTopLinksWidth = 0,
                totalBottomLinksWidth = 0,
                totalRightLinksWidth = 0,
                totalLeftLinksWidth = 0,
                totalLeftSubLinksWidth = {},
                totalRightSubLinksWidth = {};
            /*
            calculate circular link widths
             */
            graph.links.forEach(function(link) {
                if (ky != null) {
                    link.width = link.value * ky;
                } else {
                    link.width = link.width * ratio;
                }

                if (link.circular) {
                    if (link.circularLinkType == 'top') {
                        totalTopLinksWidth = totalTopLinksWidth + link.width;
                    } else {
                        totalBottomLinksWidth = totalBottomLinksWidth + link.width;
                    }

                    if (link.source.column > link.target.column) {
                        // totalLeftSubLinksWidth
                        //left side link
                        if (totalLeftSubLinksWidth[link.target.column]) {
                            totalLeftSubLinksWidth[link.target.column].total += link.width;
                        } else {
                            totalLeftSubLinksWidth[link.target.column] = {};
                            totalLeftSubLinksWidth[link.target.column].total = link.width;
                        }

                        if (totalLeftSubLinksWidth[link.target.column]) {
                            if (link.target.x0 - totalLeftSubLinksWidth[link.target.column].total < graph.x0) {
                                totalLeftSubLinksWidth[link.target.column].extra =
                                    (totalLeftSubLinksWidth[link.target.column].extra || 0) + link.width;
                            }
                        }

                        if (totalRightSubLinksWidth[link.target.column]) {
                            totalRightSubLinksWidth[link.target.column].total += link.width;
                        } else {
                            totalRightSubLinksWidth[link.target.column] = {};
                            totalRightSubLinksWidth[link.target.column].total = link.width;
                        }
                        if (totalRightSubLinksWidth[link.target.column]) {
                            if (link.target.x1 + totalRightSubLinksWidth[link.target.column].total > graph.x1) {
                                totalRightSubLinksWidth[link.target.column].extra =
                                    (totalRightSubLinksWidth[link.target.column].extra || 0) + link.width;
                            }
                        }
                    }
                    if (link.source.column < link.target.column) {
                        //right side link
                        if (totalRightSubLinksWidth[link.target.column]) {
                            totalRightSubLinksWidth[link.target.column].total += link.width;
                        } else {
                            totalRightSubLinksWidth[link.target.column] = {};
                            totalRightSubLinksWidth[link.target.column].total = link.width;
                        }
                        if (totalRightSubLinksWidth[link.target.column]) {
                            if (link.target.x1 + totalRightSubLinksWidth[link.target.column].total > graph.x1) {
                                totalRightSubLinksWidth[link.target.column].extra =
                                    (totalRightSubLinksWidth[link.target.column].extra || 0) + link.width;
                            }
                        }
                    }

                    if (link.target.column == 0) {
                        totalLeftLinksWidth = totalLeftLinksWidth + link.width;
                    }

                    if (link.source.column == maxColumn) {
                        totalRightLinksWidth = totalRightLinksWidth + link.width;
                    }
                }
            });

            let x = Object.keys(totalLeftSubLinksWidth),
                y = Object.keys(totalRightSubLinksWidth);
            (totalLeftLinksWidth +=
                d3_max(x, d => {
                    return totalLeftSubLinksWidth[d].extra || 0;
                }) || 0),
                (totalRightLinksWidth +=
                    d3_max(y, d => {
                        return totalRightSubLinksWidth[d].extra || 0;
                    }) || 0);

            //account for radius of curves and padding between links
            totalTopLinksWidth = totalTopLinksWidth > 0 ? totalTopLinksWidth + verticalMargin + baseRadius : totalTopLinksWidth;
            totalBottomLinksWidth = totalBottomLinksWidth > 0 ? totalBottomLinksWidth + verticalMargin + baseRadius : totalBottomLinksWidth;
            totalRightLinksWidth = totalRightLinksWidth > 0 ? totalRightLinksWidth + verticalMargin + baseRadius : totalRightLinksWidth;
            totalLeftLinksWidth = totalLeftLinksWidth > 0 ? totalLeftLinksWidth + verticalMargin + baseRadius : totalLeftLinksWidth;

            /*
           calculate circular link margin
            */
            var circularLinkMargin = {
                top: totalTopLinksWidth,
                bottom: totalBottomLinksWidth,
                left: totalLeftLinksWidth,
                right: totalRightLinksWidth
            };
            graph.circularSize = {
                width: circularLinkMargin.left + circularLinkMargin.right,
                height: circularLinkMargin.top + circularLinkMargin.bottom,
                hCount: x.length + y.length
            };

            // console.log("hiiiii", circularLinkMargin.top)
            return circularLinkMargin;
        }

        return graph;
    }

    return sankey;
};

/// /////////////////////////////////////////////////////////////////////////////////
// Cycle functions

//portion of code to detect circular links based on Colin Fergus' bl.ock https://gist.github.com/cfergus/3956043

// Identify circles in the link objects
function identifyCircles(graph) {
    var addedLinks = [];
    var circularLinkID = 0;
    graph.links.forEach(function(link) {
        if (createsCycle(link.source, link.target, addedLinks)) {
            graph.circular = true;
            link.circular = true;
            link.circularLinkID = circularLinkID;
            circularLinkID = circularLinkID + 1;
        } else {
            link.circular = false;
            addedLinks.push(link);
        }
    });
}

function selectCircularLinkTypes(graph) {
    if (graph.circular) {
        let numberOfTops = 0;
        let numberOfBottoms = 0;
        graph.links.forEach(function(link) {
            if (link.circular) {
                // if either souce or target has type already use that
                if (link.source.circularLinkType || link.target.circularLinkType) {
                    // default to source type if available
                    link.circularLinkType = link.source.circularLinkType ? link.source.circularLinkType : link.target.circularLinkType;
                } else {
                    link.circularLinkType = numberOfTops < numberOfBottoms ? 'top' : 'bottom';
                }

                if (link.circularLinkType == 'top') {
                    numberOfTops = numberOfTops + 1;
                } else {
                    numberOfBottoms = numberOfBottoms + 1;
                }

                graph.nodes.forEach(function(node) {
                    if (node.name == link.source.name || node.name == link.target.name) {
                        node.circularLinkType = link.circularLinkType;
                    }
                });
            }
        });
        graph.circularLinks = numberOfTops + numberOfBottoms;
    }
}

// Checks if link creates a cycle
function createsCycle(originalSource, nodeToCheck, graph) {
    if (graph.length == 0) {
        return false;
    }

    var nextLinks = findLinksOutward(nodeToCheck, graph);
    // leaf node check
    if (nextLinks.length == 0) {
        return false;
    }

    // cycle check
    for (var i = 0; i < nextLinks.length; i++) {
        var nextLink = nextLinks[i];

        if (nextLink.target === originalSource) {
            return true;
        }

        // Recurse
        if (createsCycle(originalSource, nextLink.target, graph)) {
            return true;
        }
    }

    // Exhausted all links
    return false;
}

/* Given a node, find all links for which this is a source
   in the current 'known' graph  */
function findLinksOutward(node, graph) {
    var children = [];

    for (var i = 0; i < graph.length; i++) {
        if (node == graph[i].source) {
            children.push(graph[i]);
        }
    }

    return children;
}

function horizontalSource(d) {
    return [d.source.x1, d.y0];
}

function horizontalTarget(d) {
    return [d.target.x0, d.y1];
}

function sankeyLinkHorizontal() {
    return d3_linkHorizontal.source(horizontalSource).target(horizontalTarget);
}
export {
    Sankey as sankey,
    center as sankeyCenter,
    justify as sankeyJustify,
    left as sankeyLeft,
    right as sankeyRight,
    level as sankeyLevel,
    sankeyLinkHorizontal
};

Registry.setComponent('d3_Sankey', Sankey);
