// @zohocharts/zc/src/sankey/sankey-datalabels.js
import { labelRenderer } from '../core/main-zc-init';
import { applyDatalabelFonts, isDatalabelEnabled, getLabelArray } from '../label/datalabelsUtils';
import {
    getBoundBoxHookConf,
    getBoundBoxPadding,
    getBoundBoxStrokeWidth,
    getMergedBoundBoxOptions,
    getSizeToBeAdded,
    isBoundBoxEnabled
} from '../label/boundBoxUtils';
import { getFontSizeByHierarchy } from '../utils/zcdatautils';
import { dataLabelTextContent } from '../label/datalabel-content';
import { DATALABELS_LABELTYPE } from '../label/labeltype-constants';
import { appendEle } from '../utils/zcsvgutils';
import { bbox, groupBy } from '../utils/zcutils';
import { LEFT_ALIGN, RIGHT_ALIGN } from '../core/align-constants';
import { SANKEY_CHARTNAME } from '../core/chartname-constants';
import { d3_ascending } from '../core/d3-constants';
import { translateTspan } from '../label/tspan-utils';

/**
 *
 * @return {labels}
 */
labelRenderer.sankey = function() {
    var MARGIN = 5,
        renderer,
        args;

    function labels(selection, labelHandler) {
        selection.each(function() {
            var [currentdata, index, sankeyLayout, sankeySize] = args;

            var [chartWidth, chartHeight] = sankeySize,
                paneid = labelHandler.paneid,
                rp = renderer.commonRendererProp().params,
                chartObj = rp.chartObj,
                systemconf = chartObj.systemConf,
                commonPlotOptions = rp.commonPlotOptions,
                dataset = rp.dataset,
                yaxiscolumnorder = rp.yaxiscolumnorder,
                labelArray = getLabelArray(rp, paneid),
                chartName = chartObj.chartTypes[0].name,
                stage,
                dataLabels;

            var translation = '';

            currentdata = currentdata[index];
            var plotoptions = currentdata.plotoptions,
                boundBoxPaddingAlone = 0,
                boundBoxPadding = 0,
                hookWidth = 0,
                boundBoxRadius = 0,
                { node: sankeyNodeOptions, link: sankeyLinkOptions } = commonPlotOptions;

            // dataLabels = getMergedDatalabelOptions(commonPlotOptions, plotoptions, chartObj, index);
            dataLabels = commonPlotOptions.node.datalabels;

            if (isBoundBoxEnabled(dataLabels, chartObj)) {
                var boundBoxOptions = getMergedBoundBoxOptions(dataLabels, chartObj);
                var hookOptions = getBoundBoxHookConf(boundBoxOptions.hook);
                boundBoxPaddingAlone = getBoundBoxPadding(dataLabels, chartObj);
                boundBoxPadding = boundBoxPaddingAlone + getBoundBoxStrokeWidth(dataLabels, chartObj) + hookOptions.height;
                hookWidth = hookOptions.width;
                boundBoxRadius = boundBoxOptions.radius;
            }

            var padding = dataLabels.padding || 0,
                type = dataLabels.type,
                handleOverlapping = systemconf.chart.plot.datalabels.handleOverlapping === true,
                txtHgt = getFontSizeByHierarchy(dataLabels, chartObj),
                GLT = dataLabelTextContent(chartObj, rp, dataLabels, DATALABELS_LABELTYPE, null, index, undefined, paneid),
                contentType = GLT.contentType || 'text',
                isHTMLContent = contentType === 'html',
                isMultiLine = isHTMLContent && GLT.multiLine,
                showlabel = isDatalabelEnabled(dataLabels, chartObj),
                relativeX = dataLabels.relativeX === true;

            if (!showlabel) {
                return;
            }

            showlabel = showlabel ? type : 'hidden';
            var data_length = 8,
                PathGroupPosition = [0, 0],
                idx = 0;

            var temp_label = appendEle(chartObj.staticPaneSVG.select('.labelholder'), 'text', [1], '.', 'class', 'temp_label');
            applyDatalabelFonts(temp_label, dataLabels, chartObj, rp.color);
            temp_label.attr('transform', translation).style('opacity', 0);
            var queue = labelHandler.add;

            currentdata.data.forEach(function(d, i) {
                var dataUpdate = sankeyLayout.nodes,
                    commonProperties = [];

                let alignedNodes = getAlignedNodes(sankeyLayout.nodes),
                    labelOutside = false;

                dataUpdate.forEach(function(d, j) {
                    var display = true,
                        isLeft = true,
                        heightToBeAdded = 0,
                        bb;

                    let nodeWidth = sankeyNodeOptions.width;

                    var X,
                        ANCHOR,
                        Y = d.y0 + (d.y1 - d.y0) / 2,
                        DY = isMultiLine ? '0.9em' : '0.3em';

                    if (alignedNodes.center.includes(d)) {
                        X = d.x0 + nodeWidth / 2;
                    } else if (alignedNodes.end.includes(d)) {
                        X = labelOutside ? d.x1 + 5 : d.x0 + nodeWidth / 2;
                    } else {
                        X = labelOutside ? d.x0 - 5 : d.x1 - nodeWidth / 2;
                    }

                    if (isMultiLine || alignedNodes.center.includes(d)) {
                        ANCHOR = 'middle';
                    } else if (alignedNodes.end.includes(d)) {
                        ANCHOR = labelOutside ? 'start' : 'end';
                    } else {
                        ANCHOR = labelOutside ? 'end' : 'start';
                    }

                    var text = d.name; //GLT(dt);

                    temp_label[contentType](text).attrs({
                        x: X,
                        y: Y,
                        'text-anchor': ANCHOR,
                        dy: DY
                    });

                    if (display && isMultiLine) {
                        translateTspan(temp_label, X);
                        var bb = bbox(temp_label.node());
                        Y -= bb.height / 2;
                        X += (bb.width / 2) * (isLeft ? -1 : 1);
                        translateTspan(temp_label, X);
                        temp_label.attr('y', Y);
                    }

                    if (handleOverlapping) {
                        if (display) {
                            bb = bbox(temp_label.node());
                            heightToBeAdded = getSizeToBeAdded(bb.height, boundBoxPaddingAlone, boundBoxRadius, hookWidth);
                            var DX = isLeft ? -(0 + padding + boundBoxPadding) : chartWidth + padding + boundBoxPadding;
                            if (isMultiLine) {
                                DX += (bb.width / 2) * (isLeft ? -1 : 1);
                            }

                            var obj = {
                                id: `lbl${index}_${i}_${j}`,
                                x1: bb.x - boundBoxPadding,
                                x2: bb.x + bb.width + boundBoxPadding,
                                y1: bb.y - MARGIN - boundBoxPadding - heightToBeAdded / 2,
                                isLeft: isLeft,
                                y2: bb.y + bb.height + MARGIN + boundBoxPadding + heightToBeAdded / 2,
                                DX: DX - X,
                                overlap: 'hidden',
                                translateFunction: labelHandler.update
                            };
                            labelArray.push(obj);
                        }
                    } else {
                        bb = bbox(temp_label.node());
                        heightToBeAdded = getSizeToBeAdded(bb.height, boundBoxPaddingAlone, boundBoxRadius, hookWidth);
                    }

                    var coordinate = {
                        X: X,
                        Y: Y,
                        ANCHOR: ANCHOR,
                        DY: DY,
                        text: text,
                        display: display,
                        ID: `lbl${index}_${i}_${j}`,
                        width: display ? bb.width : 0,
                        height: display ? bb.height + heightToBeAdded : 0,
                        hookPosition: isLeft ? RIGHT_ALIGN : LEFT_ALIGN
                    };

                    if (isHTMLContent) {
                        coordinate.isLeft = isLeft;
                    }

                    commonProperties.push(coordinate);

                    idx += display ? 1 : 0;
                });

                idx = -1;
                // add to drawing queue
                var q = queue(commonProperties, index, i);
                q.properties = {
                    type: SANKEY_CHARTNAME,
                    dataLabels: dataLabels,
                    dataUpdate: dataUpdate,
                    showlabel: showlabel,
                    rp: rp,
                    translation: translation,
                    trans: PathGroupPosition,
                    txtHgt: txtHgt,
                    reverseColor: true,
                    onOverlap: function() {
                        stage = 'overlap';
                    },
                    contentType: contentType,
                    isMultiLine: isMultiLine
                };

                (labelArray.TOP = -PathGroupPosition[1]), (labelArray.BOTTOM = chartHeight + PathGroupPosition[1]);
                labelArray.actualType = SANKEY_CHARTNAME;
            });
        });
    }

    function getAlignedNodes(nodes) {
        let grouped = groupBy(nodes, 'layer');
        let keys = Object.keys(grouped)
            .map(parseFloat)
            .sort(d3_ascending);

        let startKey = keys.shift();
        let endKey = keys.pop();
        let centerKey = keys;

        let centers = [];
        centerKey.forEach(key => {
            centers = [...centers, ...grouped[key]];
        });

        return {
            start: grouped[startKey],
            center: centers,
            end: grouped[endKey] || []
        };
    }

    labels.renderer = function(_) {
        if (!arguments.length) {
            return renderer;
        }
        renderer = _;
        return labels;
    };
    labels.args = function(_) {
        if (!arguments.length) {
            return args;
        }
        args = _;
        return labels;
    };
    return labels;
};

export var _sankeyDatalabel = labelRenderer.sankey;
