// @zohocharts/zc/src/sankey/sankey-chart-layout.js
//$Id$
import {
    sankey as d3_sankey,
    sankeyCenter as d3_sankeyCenter,
    sankeyJustify as d3_sankeyJustify,
    sankeyLeft as d3_sankeyLeft,
    sankeyRight as d3_sankeyRight,
    sankeyLevel as d3_sankeyLevel
} from './d3-sankey-polyfill';
import { mathMin } from '../core/zc-constants';
import { pick } from '../utils/zcutils';

const sankeyAlign = {
    left: d3_sankeyLeft,
    right: d3_sankeyRight,
    center: d3_sankeyCenter,
    justify: d3_sankeyJustify,
    level: d3_sankeyLevel
};

/**
 *
 * @param extent
 * @param iterations
 * @param sankeyOptions
 * @return {*}
 */
export function getSankeyLayout(sankeyOptions, extent, iterations, nodeMaxWidth = 0, margin, chartObj) {
    let { link: linkOptions } = sankeyOptions;
    let layout = d3_sankey()
        .extent(extent)
        .nodeWidth(mathMin(nodeMaxWidth, sankeyOptions.node.width))
        .nodeSort(sankeyOptions.node.sort)
        .linkSort(sankeyOptions.link.sort)
        .nodePadding(sankeyOptions.node.padding || 0)
        .linkPadding(linkOptions.padding || 0)
        .iterations(pick(iterations, sankeyOptions.iterations, 64))
        .nodeId(node => node.name)
        .height(extent[1][1])
        .margin(margin)
        .chart(chartObj);

    if (sankeyOptions.align) {
        layout.nodeAlign(sankeyAlign[sankeyOptions.align]);
    }

    return layout;
}
