// @zohocharts/zc/src/sankey/sankey-svg.js
import { createElement } from '../utils/zcsvgutils';
import { d3_path } from '../core/d3-constants';
import { drawSankeyNode } from './sankey-node';
import { sankeyLinkBezier, sankeyLinkLinear } from './sankey-link';
import { drawSankeyExitLink } from './sankey-exit-link';
import { defined } from '../utils/zcutils';

/**
 *
 * @param parent
 * @param layoutData
 * @param layout
 * @param nodeOptions
 * @param linkOptions
 * @param nodeFillColor
 * @param linkFillColor
 */
export function drawSankeySvg(parent, layoutData, layout, nodeOptions, linkOptions, nodeFillColor, linkFillColor) {
    let { nodes, links } = layoutData;
    
    let sankeyChart = createElement('g', '.sankey-chart', [1], parent).attr(
            'transform',
            `translate(${nodeOptions.strokeWidth},${nodeOptions.strokeWidth})`
        ),
        sankeyChart_main = createElement('g', '.sankey-chart-main', [1], sankeyChart),
        linkParent = createElement('g', '.sankey-links', [1], sankeyChart_main),
        nodeParent = createElement('g', '.sankey-nodes', [1], sankeyChart_main);

    if(layout.zoomTransform){
        let t = layout.zoomTransform;
        sankeyChart_main.attr("transform", 'translate(' + parseFloat(t.x) + ',' + parseFloat(t.y) + ') scale(' + parseFloat(t.k) + ',' + parseFloat(t.k) + ')')
    }

    createElement('path', '.sankey-link', links, linkParent, null, p => p.point)
        .attrs({
            d: function(link) {
                this.index = link.index;
                return getSankeyPointPath(link, nodeOptions, linkOptions, 0.5, layoutData);
            }
        })
        .styles({
            fill: link => linkFillColor(link.point),
            'fill-opacity': linkOptions.fillOpacity,
            stroke: linkOptions.strokeColor,
            'stroke-width': linkOptions.strokeWidth,
            'mix-blend-mode': linkOptions.blendMode
        });

    createElement('path', '.sankey-node', nodes, nodeParent, null, p => p.point)
        .attrs({
            d: node => getSankeyPointPath(node, nodeOptions, linkOptions)
        })
        .styles({
            fill: node => nodeFillColor(node.point),
            'fill-opacity': nodeOptions.fillOpacity,
            stroke: nodeOptions.strokeColor,
            'stroke-width': nodeOptions.strokeWidth
        });
}

/**
 *
 * @param point
 * @param nodeOptions
 * @param linkOptions
 * @param minWidth
 * @return {Path}
 */
export function getSankeyPointPath(point, nodeOptions, linkOptions, minWidth = 0.5, layout) {
    minWidth = defined(linkOptions.minWidth) ? linkOptions.minWidth : minWidth;
    let path = d3_path();
    if (point._type === 'node' || point._type === 'node') {
        drawSankeyNode(path, point, nodeOptions);
    } else if (point._type === 'link') {
        if (point.circular) {
            path = point.circularPathData.path;
        } else if (point.exit) {
            drawSankeyExitLink(path, point, linkOptions, minWidth, layout);
        } else {
            (linkOptions.curve === 'linear' ? sankeyLinkLinear : sankeyLinkBezier)(path, point, linkOptions, minWidth);
        }
    }
    return path;
}
