// @zohocharts/zc/src/sankey/SankeyHighlighter.js
import { defined, getPropVal, inherit, invokeFunction, isAncestorDisabled, isArray, pick, simpleClone } from '../utils/zcutils';
import { patternFillEffect } from '../events/visualEffects/highlight/pattern-fill';
import { HIGHLIGHT_CONTEXT } from '../events/visualEffects/highlight/highlight-context';
import { createElement, pathToPolygon } from '../utils/zcsvgutils';
import { ScatterHighlighter } from '../scatter/scatterHighlight';
import { getSankeyPointPath } from './sankey-svg';
import { isWithinColorRange } from '../utils/colorutils';
import { d3_polygonCentroid } from '../core/d3-constants';
import { mathAbs } from '../core/zc-constants';
import { sankeyGradientDimension } from './sankey-chart';

export function SankeyHighlighter(chart, renderer) {
    this.chart = chart;
    this.renderer = renderer;
}

inherit(SankeyHighlighter, ScatterHighlighter);

var proto = SankeyHighlighter.prototype;

proto._defaultH = function() {
    return [
        {
            shape: 'poly',
            path: this.path,
            data: this.data,
            context: this.renderingContext,
            translate: this.translateContext,
            styles: point => {
                let nodeIndex;
                if (point._type === 'link') {
                    nodeIndex = this.renderer.getSankeyNodeByPoint(point).index;
                } else {
                    nodeIndex = point.itemIndex;
                }
                return patternFillEffect.call(this, point, 'inherit', nodeIndex);
            }
        }
    ];
};

proto.data = function(point) {
    var self = this,
        chart = this.chart,
        // priority for event. less number means more priority
        rank = { node: 0, link: 1 };

    /**
     *
     * @param type
     * @return {*}
     * @private
     */
    // configurable polygon type
    point._setPolygonType = type => (point._polygonType = type);
    point._polydata = this.poly(point);
    // point._getDependentsPoints = point => self.getDependentLayoutPoints(point, true).map(d => d.point);
    point._getRank = () => rank[point._type];

    return function() {
        if (!chart.cache.data[point._key]) {
            chart.cache.data[point._key] = {};
        }

        var cache = chart.cache.data[point._key];
        if (!defined(cache._x) && (!chart.dataObject.hasHierarchyData || !isAncestorDisabled(chart.seriesdata, point))) {   //Condition to retain disabled items in legend
            let polygonPoints = self.getPolygon(point, true);
            let [x, y] = d3_polygonCentroid(polygonPoints);
            cache._x = point._x = x;
            cache._y = point._y = y;
            cache._defined = point._defined = true;
            cache._point = point;
            cache._poly = point._poly = { points: polygonPoints };  //* caching polypoints for sankey highlight performance
        }
        cache._visible = point._visible = self.isVisible(point);

        return point;
    };
};

proto.isVisible = function(point, from) {
    var self = this,
        chart = self.chart,
        disabledItems = chart.cache.disabledItems,
        nodeName = chart.dataset.getX(point);

    let source = isArray(nodeName) ? nodeName[0] : nodeName,
        target = isArray(nodeName) ? nodeName[1] : chart.dataset.getY(point, 0),
        weight = chart.dataset.getZ(point),
        negativeAsAbsolute = getPropVal(chart.systemConf, `chart.plot.plotoptions.sankey.showAsNegative`) === 'absolute',
        hasFixedValue = false;

    // #ZC2589
    if (negativeAsAbsolute) {
        weight = mathAbs(weight);
    }

    let visible = !disabledItems.has(source) && !disabledItems.has(target) && weight > 0;

    let colorRangeCheck = ((point._type === 'node' || point._type === 'link') ? 
        (chart.systemConf.legend.ranges) ? 
            (isWithinColorRange(point[0], chart)) && ((point[1] != null) ? (isWithinColorRange(point[1], chart)) : true) :
            (isWithinColorRange(chart.dataset.getClr(point), chart)) :
        false);

    // fixed values is applied when no link is attached after legend filter
    if (visible && point._type === 'node' && from !== 'legend') {
        const node = self.renderer.getSankeyNodeByPoint(point);
        hasFixedValue = node ? node.hasOwnProperty('fixedValue') : hasFixedValue;
    }

    return visible && colorRangeCheck && !point.disabled && !hasFixedValue;
};

/**
 *
 * @param point
 * @return {*[]}
 */
proto.getPolygon = function(point, zoom) {
    let self = this,
        renderer = self.renderer,
        polygon,
        { link: linkOptions, events = {} } = renderer.sankeyOptions,
        sampleSize = pick(events.sample, 50),
        isBezierCurve = linkOptions.curve === 'bezier';

    /**
     * sankey point has 2 types
     * 1. polygon
     * 2. path -- needs conversion
     */
     let zoom_transform = self.chart.dataObject.sankeyLayout.zoomTransform;
    if (point._type === 'link') {
        let link = renderer.getLinkByPoint(point);

        // ZC2589
        if (link) {
            if(!zoom_transform || !link._polypointsOrig){
                if (point._polygonType === 'path' || isBezierCurve || link.circular) {
                    let path = renderer.getSankeyPath(link, 4);
                    polygon = pathToPolygon(path, sampleSize);
                } else {
                    polygon = link._polypoints;
                }
                link._polypoints = polygon;
                var isZoom = self.chart.systemConf.chart.zoom.enabled;
                if(isZoom && !link._zoomed){
                    link._polypointsOrig = simpleClone(polygon);
                }
            }
            if(zoom_transform){
                polygon = link._polypointsOrig;
                let polygon_orig = simpleClone(polygon);
                link._polypoints = polygon = polygon_orig.map(d => {
                    d[0] = (d[0] * parseFloat(zoom_transform.k)) + parseFloat(zoom_transform.x);
                    d[1] = (d[1] * parseFloat(zoom_transform.k)) + parseFloat(zoom_transform.y);
                    return d;
                })
                link._zoomed = true;
            }
        }
    } else {
        let node = renderer.getSankeyNodeByPoint(point);
        if (node) {
            // node and liner link has pre-calculated polygon while rendering
            polygon = node._polypoints;

            if(zoom_transform && zoom){
                if(!node._zoomed){
                    node._polypointsOrig = simpleClone(node._polypoints);
                }
                let polygon_orig = simpleClone(node._polypointsOrig);
                
                node._polypoints = polygon = polygon_orig.map(d => {
                    d[0] = (d[0] * parseFloat(zoom_transform.k)) + parseFloat(zoom_transform.x);
                    d[1] = (d[1] * parseFloat(zoom_transform.k)) + parseFloat(zoom_transform.y);
                    return d;
                })
                node._zoomed = true;
            }
        }
    }
    return polygon || [[null, null]];
};

/**
 *
 * @param point
 * @return {function(): {points: *[]}}
 */
proto.poly = function(point) {
    var self = this,
        chart = self.chart;

    return function() {
        if (!chart.cache.data[point._key]) {
            chart.cache.data[point._key] = {};
        }
        var cache = chart.cache.data[point._key];
        if (!cache._poly) {
            // get the polygon
            cache._poly = point._poly = { points: self.getPolygon(point) };
        }
        return point._poly;
    };
};

proto.renderingContext = function() {
    return HIGHLIGHT_CONTEXT.get('area').call(this, this.chart);
};

proto.translateContext = function() {
    return [this.chart.plotarea.left, this.chart.plotarea.top];
};

proto.options = function(point) {
    var rp = this.renderer.commonRendererProp().params,
        commonOptions = rp.commonPlotOptions,
        { fillOpacity } = rp.hoverOptions.fillOpacity;

    commonOptions.fillOpacity = fillOpacity;
    if (commonOptions[point._type].fillColor) {
        const link = point._type === 'link' ? this.renderer.getLinkByPoint(point) : this.renderer.getSankeyNodeByPoint(point);  // #ZC4727
        commonOptions.color = invokeFunction(commonOptions[point._type].fillColor, link, link.index, link.point.arcColor, this.chart)
    } else {
        commonOptions.color = this.renderer.getColor(point, true);
    }
    if (!commonOptions.link.gradients.customGradient && defined(commonOptions.link.gradients.type)) {
        sankeyGradientDimension(point, commonOptions.link.gradients);
    }

    return commonOptions;
};

proto.path = function(point, model) {
    var self = this,
        context = model.context.call(self),
        translate = model.translate.call(self),
        renderer = self.renderer,
        {
            sankeyOptions: { link: linkOptions, node: nodeOptions }
        } = renderer,
        nodeStrokeWidth = nodeOptions.strokeWidth;

    context.style('display', '').attr('transform', `translate(${translate[0] + nodeStrokeWidth},${translate[1] + nodeStrokeWidth})`);

    let layoutPoint;
    if (point._type === 'link') {
        layoutPoint = renderer.getLinkByPoint(point);
    } else if (point._type === 'node') {
        layoutPoint = renderer.getSankeyNodeByPoint(point);
    }

    let __path = createElement('path', `.sankey-highlight-${point._type}-${layoutPoint.index}`, [layoutPoint], context).attr('d', point =>
    getSankeyPointPath(point, nodeOptions, linkOptions, 0.5, renderer.sankeyLayout));

    let transform = self.chart.dataObject.sankeyLayout.zoomTransform;

    if(transform){
        __path.attr('transform', 'translate(' + transform.x + ',' + transform.y + ') scale(' + transform.k + ' ' + transform.k + ')');;
    }
    return __path;
    
};

/**
 *
 * @param point
 * @param flatten
 * @param from
 * @return
 */
proto.getDependentLayoutPoints = function(point, flatten, from) {
    let self = this,
        chart = self.chart,
        renderer = self.renderer,
        layout = self.chart.dataObject.sankeyLayout,
        { phase } = self.chart,
        filter = points => {
            return points.filter(p => {
                let { point } = p;
                if (point && point.hasOwnProperty('_visible') && phase !== 'legendFilter' && from !== 'legend') {
                    return point._visible;
                } else if (point) {
                    return (point._visible = self.isVisible(point, from));
                }
            });
        },
        layoutPoint = {
            link: renderer.getLinkByPoint,
            node: renderer.getSankeyNodeByPoint
        }[point._type](point);

    if (layoutPoint) {
        let connectedPoints = layout.getAllConnectedPoints(layoutPoint, flatten);

        // if (chart.systemConf.legend.ranges) {
        //     if (flatten) {  // connectedPoints filter for svg mode
        //         connectedPoints = connectedPoints.filter(currentLayoutPoint => (currentLayoutPoint != "null") && currentLayoutPoint.point._visible)
        //     } else { // connectedPoints filter for canvas mode
        //         connectedPoints.links = connectedPoints.links.filter(link => (link != "null") && link.point._visible)
        //         connectedPoints.nodes = connectedPoints.nodes.filter(node => (node != "null") && node.point._visible)
        //     }
        //     return connectedPoints;
        // }

        if (flatten) {
            return filter(connectedPoints);
        } else {
            return {
                links: filter(connectedPoints.links),
                nodes: filter(connectedPoints.nodes)
            };
        }
    }

    return [];
};
