// @zohocharts/zc/src/sankey/sankey-chart.js
import { renderer_helpers, svg_plot_renderer, canvas_plot_renderer } from '../core/main-zc-init';
import { d3_ascending, d3_map, d3_namespaces } from '../core/d3-constants';
import { arrayIncludes, defined, groupBy, invokeFunction, mergeJSON, pick } from '../utils/zcutils';
import { getFillColor, getRawColor, isWithinColorRange } from '../utils/colorutils';
import { mathMin, NULL, UNDEFINED } from '../core/zc-constants';
import { drawSankeySvg, getSankeyPointPath } from './sankey-svg';
import { drawSankeyCanvas } from './sankey-canvas';
import { getSankeyLayout } from './sankey-chart-layout';
import { END_ALIGN, MIDDLE_ALIGN, START_ALIGN } from '../core/align-constants';
import { SankeyHighlighter } from './SankeyHighlighter';
import { svg_clips } from '../effects/svg-clips';
import { getAnimationConf } from '../utils/animationutils';
import { SANKEY_CHARTNAME } from '../core/chartname-constants';
import { LabelRegistry } from '../label/label-registry';
import { getBoundBoxHookConf, getMergedBoundBoxOptions } from '../label/boundBoxUtils';
import { getFormattedValue } from '../formatter/formatter';
import { Registry } from '../core/zc-registry';

renderer_helpers.sankey = {};
function SankeyChart() {
    let commonRendererProp,
        sankeyLabelInside = false,
        sankeyNodeAligns,
        sankeyLayoutData,
        nodeMapped,
        linkMapped,
        chartInstance,
        nodeFillColor,
        linkFillColor,
        linkPadding,
        sankeyHierarchyLayout,
        labelOutside = false,
        labelWidthStart = 0,
        labelWidthEnd = 0,
        sankeyOptions,
        tempPath,
        /**
         *
         */
        chart = function(selection, data, styles, filterIndex, encode, itemIndex) {
            chartInstance = commonRendererProp.params.chartObj;

            let params = commonRendererProp.params,
                { labelQueue } = params,
                commonPlotOptions = params.commonPlotOptions,
                systemConf = chartInstance.systemConf,
                dataset = params.dataset,
                dataObject = chartInstance.dataObject,
                index = params.datasetIndex || 0,
                series = chartInstance.seriesdata[index],
                isCanvasMode = chartInstance.plot.renderer.mode === 'canvas',
                plotoptions = series.plotoptions;

            sankeyOptions = mergeJSON(pick(series.plotoptions, {}), commonPlotOptions);
            sankeyHierarchyLayout = dataObject.sankeyHierarchyLayout;

            let { link: linkOptions, node: nodeOptions } = sankeyOptions,
                { legendItemsMap, legend } = chartInstance.cache.sankey,
                { graph: layout, linkLength, nodeLength } = dataObject.sankeyLayout,
                sankeyLayout = dataObject.sankeyLayout,
                { width, height } = chartInstance.plotarea,
                linkGradientOptions = linkOptions.gradients,
                linearDim = linkGradientOptions.options.linear,
                customGradient = defined(linearDim.x1) || defined(linearDim.x2) || defined(linearDim.y1) || defined(linearDim.y2),
                isRedraw = defined(chartInstance.phase),
                isCanvasLegendHighlight = isCanvasMode && defined(itemIndex),
                animationOptions = getAnimationConf(commonPlotOptions, plotoptions, chartInstance, SANKEY_CHARTNAME),
                // #ZC2610
                animationDuration = isRedraw ? 0 : animationOptions.duration,
                nodeStrokeWidth = nodeOptions.strokeWidth,
                labelOptions = sankeyOptions.datalabels,
                labelPadding = 0,
                hookHeight = 0,
                imagePallet = chartInstance.imagePallete,
                eventOptions = mergeJSON(pick(systemConf.chart.plot.events), commonPlotOptions.events),
                boundBoxPaddingSingle = 0,
                isLegendHighlight = defined(itemIndex),
                // TODO revert to $sankeyAlignMinSize = 1
                // ZC2613 The issue fixes if $sankeyAlignMinSize = 1. Since it causes image diffs. We will do it in next milestone
                sankeyAlignMinSize = 0;

            const hasRanges = defined(systemConf.legend.ranges);

            customGradient = defined(linkGradientOptions.customGradient) ? linkGradientOptions.customGradient : customGradient;
            linkGradientOptions.customGradient = customGradient;
            if (chartInstance.systemConf.chart.width) {
                width = invokeFunction(chartInstance.systemConf.chart.width, chartInstance);
            }
            if (chartInstance.systemConf.chart.height) {
                height = invokeFunction(chartInstance.systemConf.chart.height, chartInstance);
            }

            // get the node aligns
            sankeyNodeAligns = chart.nodeAligns = getAlignedNodes(layout.nodes);
            sankeyLabelInside = labelOptions.labelPos === 'inside';
            labelOutside = labelOptions.labelPos === 'outside';
            linkPadding = linkOptions.padding || 0;

            // create maps for access
            nodeMapped = d3_map();
            
            layout.nodes.forEach(function(node) {
                nodeMapped.set(node.name, node);
                node.point.arcColor = sankeyHierarchyLayout && node.color ? node.color
                    : getRawColor(chartInstance, node.index, node.point, index);
                (hasRanges) ? node.point._visible = isWithinColorRange(node.point[0], chartInstance) : null;
            });

            if (hasRanges) {
                sankeyLayout.nodeMatrix = Array.from({ length: sankeyLayout.layout.columns}, () => []);
                const nodeMatrix = sankeyLayout.nodeMatrix;
                layout.nodes.forEach( d => nodeMatrix[d.column].push(d))
            }

            linkMapped = d3_map();
            layout.links.forEach(function(link) {
                linkMapped.set(link.point._key, link);
                (hasRanges) ? link.point._visible = (isWithinColorRange(link.point[0], chartInstance) && ((link.target != "null") ? isWithinColorRange(link.point[1], chartInstance): true)) : null;

                if (linkGradientOptions.type) {
                    if (link.circular) {
                        link.point.arcColor = [link.target.point.arcColor, link.source.point.arcColor];
                    } else if (link.exit) {
                        link.point.arcColor = linkOptions.exit.color;
                        link.point.exit = true;
                    } else {
                        link.point.arcColor = [link.source.point.arcColor, link.target.point.arcColor];
                    }
                } else {
                    link.point.arcColor = link.exit ? linkOptions.exit.color
                        : sankeyHierarchyLayout && chartInstance.dataset.clridx >= 0 ? getRawColor(chartInstance, link.index, link.point, index)
                            : link.color || link.source.point.arcColor;
                }
            });

            // disabled because canvas highlight issue. and it is correct
            // legend = legend.filter(d => !d.disabled);

            // pre-calculate the edge labels and reduce that from sankey area.
            // so that we can place the label outside
            if (labelOutside && legend.length) {
                let boundBoxOptions = getMergedBoundBoxOptions(labelOptions, chartInstance);
                let hookOptions = getBoundBoxHookConf(boundBoxOptions.hook),
                    boundBoxPadding = (boundBoxOptions.padding || 0) * 2,
                    nodeMaxWidth = (width / sankeyNodeAligns.layerCount) * 0.8;

                hookHeight = hookOptions.height;
                labelPadding = (labelOptions.padding || 0) * 2 + boundBoxPadding;
                boundBoxPaddingSingle = boundBoxPadding / 2;

                // let tempExtent = [
                //     [labelWidthStart + nodeStrokeWidth + labelPadding + hookHeight, nodeStrokeWidth + boundBoxPadding],
                //     [width - nodeStrokeWidth - labelWidthEnd - labelPadding - hookHeight, height - nodeStrokeWidth - boundBoxPadding]
                // ];
                //
                // let rawSankeyLayout = getSankeyLayout(sankeyOptions, tempExtent, UNDEFINED, nodeMaxWidth);
                // layout = linkLength && nodeLength ? rawSankeyLayout(layout) : layout;

                // ZC2609
                if (isLegendHighlight) {
                    [labelWidthStart, labelWidthEnd] = chartInstance.renderer[0].labelWidthOutside;
                } else {
                    [labelWidthStart, labelWidthEnd] = LabelRegistry.get('preCalculation')(null, {}, chartInstance) || [0, 0];
                    chartInstance.renderer[0].labelWidthOutside = [labelWidthStart, labelWidthEnd];
                }
            }

            let leftMargin = labelWidthStart + nodeStrokeWidth + labelPadding + hookHeight,
                rightMargin = labelWidthEnd + nodeStrokeWidth + labelPadding + hookHeight,
                verticalMargin = nodeStrokeWidth + boundBoxPaddingSingle + sankeyAlignMinSize,
                sankeyExtent = [[leftMargin, verticalMargin], [width - rightMargin, height - verticalMargin]],
                gradientDim = [width, height],
                nodeFillColorRaw = getFillColor(chartInstance, 0),
                linkFillColorRaw = getFillColor(chartInstance, 0, null, gradientDim, null, null, linkGradientOptions),
                hoverOptions = (commonRendererProp.params.hoverOptions =
                    plotoptions && plotoptions.hoveroptions
                        ? mergeJSON(commonPlotOptions.hoveroptions, plotoptions.hoveroptions, true)
                        : commonPlotOptions.hoveroptions);

            let availablePlotWidth = sankeyExtent[1][0] - sankeyExtent[0][0];
            // sankey max node with is = 80% width / layer
            let maxNodeWidth = (availablePlotWidth / sankeyNodeAligns.layerCount) * 0.8;

            /**
             *
             * @param point
             * @param calculate
             * @return {*}
             */
            nodeFillColor = function(point, calculate) {
                let node = getSankeyNodeByPoint(point),
                    itemIndex;

                const fillColor = invokeFunction(nodeOptions.fillColor, node, node.index, node.point.arcColor, chartInstance);

                if (!node) {
                    return;
                }

                itemIndex = legendItemsMap.get(node.name).index;

                if (calculate) {
                    return nodeFillColorRaw(itemIndex, node.point, NULL, NULL, nodeOptions.fillOpacity);
                }
                return nodeFillColorRaw(itemIndex, node.point, NULL, fillColor, nodeOptions.fillOpacity);
            };

            /**     *
             * @param point
             * @return {*}
             */
            linkFillColor = function(point) {
                let link = linkMapped.get(point._key),
                    { fillOpacity } = isCanvasLegendHighlight ? hoverOptions : linkOptions;

                if (hasRanges && link.point._visible === false) {
                    const legend = chartInstance.systemConf.legend,
                        filterTrackfillColor = legend.filter && legend.filter.track && legend.filter.track.fillColor
                    return (filterTrackfillColor) ? filterTrackfillColor : "rgb(237, 234, 234)";
                }

                if (defined(linkGradientOptions.type)) {
                    let { source, target } = link,
                        fillColor = link.exit ? linkOptions.exit.color : invokeFunction(linkOptions.fillColor, link, link.index, link.point.arcColor, chartInstance),
                        sourceColor = link.exit
                            ? linkOptions.exit.color[0]
                            : link.circular
                            ? nodeFillColor(target.point, true)
                            : nodeFillColor(source.point, true),
                        targetColor = link.exit
                            ? linkOptions.exit.color[1]
                            : link.circular
                            ? nodeFillColor(source.point, true)
                            : nodeFillColor(target.point, true),
                        dim = link.exit ? [source.x1 + link.width, 500] : [target.x0 - source.x1, 500];
                    if (!customGradient) {
                        sankeyGradientDimension(point, linkGradientOptions);
                    }

                    linkGradientOptions.options.linear.stopColor = fillColor || [sourceColor, targetColor];

                    return linkFillColorRaw(source.index, link.point, null, fillColor, fillOpacity, dim);
                } else {
                    let sourceIndex = legendItemsMap.get(link.source.name).index,
                        fillColor = link.exit ? linkOptions.exit.color : invokeFunction(linkOptions.fillColor, link, link.index, link.point.arcColor, chartInstance);

                    return linkFillColorRaw(sourceIndex, link.point, NULL, fillColor, fillOpacity, null, {
                        context: chartInstance.context
                    });
                }
            };
            let _layout = chartInstance.dataObject.sankeyLayout.layout,
                exitAtEndNode = _layout  && _layout.exitAtEndNode;
            if (exitAtEndNode) {
                sankeyExtent[1][0] = sankeyExtent[1][0] - mathMin(exitAtEndNode, 25);
            }
            let rawSankeyLayout = getSankeyLayout(sankeyOptions, sankeyExtent, UNDEFINED, maxNodeWidth, {
                left: leftMargin,
                right: rightMargin
            }, chartInstance);

            chart.sankeyLayout = sankeyLayoutData = linkLength || nodeLength ? rawSankeyLayout(layout) : layout;
            dataObject.sankeyLayout.layout = chart.sankeyLayout;

            sankeyLayoutData.matrix?.nodes.forEach( (column) => column.forEach((d, i) => d.sortedIndex = i));  // sortedIndex for tints

            // update index
            if (chartInstance.phase === 'legendFilter') {
                sankeyLayoutData.nodes.forEach(node => {
                    node.index = legendItemsMap.get(node.name).index;
                });
            }

            // chart.seriesdataAfterUpdate = [legend];
            chart.seriesdataAfterUpdate = [sankeyLayoutData.nodes.map(_ => _.point)];
            chart.sankeyOptions = sankeyOptions;

            if (isCanvasMode) {
                // for legend highlight
                if (isLegendHighlight) {
                    itemIndex = (chartInstance.chartTypes[0].name === SANKEY_CHARTNAME && chartInstance.systemConf.legend.ranges) ? getSankeyNodeByPoint(chartInstance.systemConf.legend.ranges[itemIndex]).point.itemIndex : itemIndex;
                    let highlighter = chartInstance.renderer[0].highlighter,
                        currentSelectedPoint = legend[itemIndex],
                        layoutPoints = highlighter.getDependentLayoutPoints(currentSelectedPoint),
                        // #ZC2579 [$linkHoverStyles, $nodeHoverStyles]
                        linkHoverStyles = { ...linkOptions, fillColor: styles.fill, fillOpacity: styles['fill-opacity'] },
                        nodeHoverStyles = { ...nodeOptions, fillColor: styles.fill };
                    layoutPoints.x0 = sankeyLayoutData.x0;
                    layoutPoints.x1 = sankeyLayoutData.x1;
                    layoutPoints.y0 = sankeyLayoutData.y0;
                    layoutPoints.y1 = sankeyLayoutData.y1;

                    drawSankeyCanvas(selection, layoutPoints, nodeHoverStyles, linkHoverStyles, nodeFillColor, linkFillColor, imagePallet, chartInstance);
                } else {
                    drawSankeyCanvas(selection, sankeyLayoutData, nodeOptions, linkOptions, nodeFillColor, linkFillColor, imagePallet, chartInstance);
                }
            } else {
                if (!isRedraw) {
                    animateSankeySvg(selection, animationOptions);
                } else {
                    updateAnimatorClipPath(selection);
                }

                drawSankeySvg(selection, sankeyLayoutData, chartInstance.dataObject.sankeyLayout, nodeOptions, linkOptions, nodeFillColor, linkFillColor);
            }

            let labelPoints = linkLength || nodeLength ? legend.filter(d => !d.disabled) : [];

            labelQueue.length = 0;
            labelQueue.push({
                renderer: chart,
                charttype: 'hierarchy',
                chartObj: chartInstance,
                args: [chartInstance.seriesdata[0], 0, NULL, NULL, null, labelPoints, [0, 0]]
            });

            // create a path element for flatten the bezier curve to linear
            tempPath = document.createElementNS(d3_namespaces.svg, 'path');

            // for highlighter
            chart.getPath = (point, minWidth) => {
                tempPath.setAttribute('d', getSankeyPointPath(point, nodeOptions, linkOptions, minWidth, sankeyLayoutData));
                return tempPath;
            };
            chart.getSankeyPath = (point, minWidth) => {
                const path = getSankeyPointPath(point, nodeOptions, linkOptions, minWidth, sankeyLayoutData);
                return path._ || path;
            };

            // method to get link by the actual series point
            chart.getLinkByPoint = point => linkMapped.get(point._key) || point._root && point._root.link;
            chart.getNodeByPoint = point => nodeMapped.get(point._key);

            // #ZC2588
            legend.disabled = () => eventOptions.enabled === false;
            chartInstance.cache.customPoints.sankeyNodes = { '0_0': legend };

            //
            setTimeout(function() {
                chart.chartRendered = true;
            }, animationDuration);

            // tempPath = null;    // memory leak
        };

    /**
     *
     * @param nodes
     * @return {{center: [], start: *, end: (*|*[])}}
     */
    function getAlignedNodes(nodes) {
        let grouped = groupBy(nodes, 'layer');
        let keys = Object.keys(grouped)
                .map(parseFloat)
                .sort(d3_ascending),
            layerCount = keys.length;

        let startKey = keys.shift();
        let endKey = keys.pop();
        let centerKey = keys;

        let centers = [];
        centerKey.forEach(key => {
            centers = [...centers, ...grouped[key]];
        });

        return {
            start: grouped[startKey],
            center: centers,
            end: grouped[endKey] || [],
            layerCount
        };
    }

    function getSankeyNodeByPoint(point, customNodeMap) {
        const value = chartInstance.dataObject.isLevelAxisCategory ? point[chartInstance.dataset.levelidx(-1)] : chartInstance.dataset.getX(point);
        let key = point._type == 'link' ?
            sankeyHierarchyLayout ? sankeyHierarchyLayout.accessor('source')(point)
                : formatPoint(value, point._format || 'x') : value;
        return pick(customNodeMap, nodeMapped).get(key);
    }

    // function getSankeyLinkByPoint(point, customLinkMap) {
    //     let key = point._key;
    //     return pick(customLinkMap, linkMapped).get(key);
    // }

    /**
     *
     * @param d
     * @param computedLabelBound
     * @param padding
     * @param maxWidth
     * @param boundBoxPadding
     * @param hookHeight
     * @return {{x: number, width: number, y: number, height: number}}
     */
    function getSankeyLabelPosition(d, computedLabelBound, padding = 0, maxWidth, boundBoxPadding = 0, hookHeight = 0) {
        let node = getSankeyNodeByPoint(d);

        if (sankeyLabelInside) {
            return {
                x: node.x0,
                y: node.y0,
                width: node.x1 - node.x0,
                height: node.y1 - node.y0
            };
        } else {
            let x,
                nodeWidth = node.x1 - node.x0,
                nodeHeight = node.y1 - node.y0,
                labelWidth = 0,
                bPadding = boundBoxPadding * 2;

            if (computedLabelBound) {
                labelWidth = mathMin(computedLabelBound.width, maxWidth);
            }

            if (arrayIncludes(sankeyNodeAligns.start, node)) {
                if (labelOutside) {
                    x = node.x0 - hookHeight - padding * 2 - labelWidth - bPadding;
                } else {
                    x = node.x1 + hookHeight;
                }
            } else if (arrayIncludes(sankeyNodeAligns.end, node)) {
                if (labelOutside) {
                    x = node.x1 + hookHeight;
                } else {
                    x = node.x0 - labelWidth - hookHeight - padding * 2 - bPadding;
                }
            } else {
                let nodeCenter = node.x0 + nodeWidth / 2;
                x = nodeCenter - labelWidth / 2 - padding - bPadding;
            }

            return {
                x: x,
                y: node.y0,
                width: labelWidth + padding * 2 + bPadding,
                height: node.y1 - node.y0
            };
        }
    }

    /**
     *
     * @param d
     * @return {string}
     */
    function getLabelAnchor(d) {
        let node = getSankeyNodeByPoint(d),
            anchor;

        // return START_ALIGN;
        if (sankeyLabelInside) {
            return MIDDLE_ALIGN;
        } else {
            if (arrayIncludes(sankeyNodeAligns.center, node)) {
                anchor = MIDDLE_ALIGN;
            } else if (arrayIncludes(sankeyNodeAligns.end, node)) {
                anchor = labelOutside ? START_ALIGN : END_ALIGN;
            } else {
                anchor = labelOutside ? END_ALIGN : START_ALIGN;
            }
            return anchor;
        }
    }

    /**
     *
     * @param selection
     * @param animationOptions
     */
    function animateSankeySvg(selection, animationOptions) {
        let clipDimension = { x: 0, y: 0, width: 0, height: chartInstance.plotarea.height },
            clipRect = updateAnimatorClipPath(selection, clipDimension);

        clipRect
            .transition()
            .ease(animationOptions.easingType)
            .duration(animationOptions.duration)
            .attr('width', chartInstance.plotarea.width);
    }

    /**
     *
     * @param selection
     * @param clipDimension
     */
    function updateAnimatorClipPath(
        selection,
        clipDimension = { x: 0, y: 0, width: chartInstance.plotarea.width, height: chartInstance.plotarea.height }
    ) {
        return sankeyClipPath(chartInstance, selection, clipDimension);
    }

    function formatPoint(value, type) {
        var column = chartInstance.dataset.getColumn(type, 0);
        return getFormattedValue(column[column.datatype], value, column.datatype, chartInstance);
    }

    /**
     *
     * @param chartObj
     * @param index
     * @return {SankeyHighlighter}
     */
    chart.getHighlighter = function(chartObj, index) {
        return (chart.highlighter = new SankeyHighlighter(chartObj, chart, sankeyLayoutData));
    };

    chart.getPosition = getSankeyLabelPosition;
    chart.getLabelAnchor = getLabelAnchor;
    chart.getSankeyNodeByPoint = getSankeyNodeByPoint;
    // chart.getSankeyLinkByPoint = getSankeyLinkByPoint;

    /**
     *
     * @param point
     * @param isHighlight
     * @return {[*, *]|string|*}
     */
    chart.getColor = function(point, isHighlight) {
        let customColor = sankeyOptions[point._type].fillColor;

        if (!defined(customColor) && defined(point.arcColor)) {
            return point.arcColor;
        }

        if (point._type === 'node') {
            return nodeFillColor(point);
        } else if (point._type === 'link') {
            return linkFillColor(point, isHighlight);
        }
    };

    /**
     *
     * @param _
     * @return {*}
     */
    chart.commonRendererProp = _ => (!_ ? commonRendererProp : ((commonRendererProp = _), chart));

    chart.destroy = function () {
        tempPath = null;
    }

    return chart;
}

/**
 *
 * @param chartInstance
 * @return {number|*}
 */
function isPatternEnabled(chartInstance) {
    let patternPallets = chartInstance.patternPallete,
        imagePallets = chartInstance.imagePallete;

    return (imagePallets && imagePallets.length) || (patternPallets && patternPallets.length);
}

export function sankeyClipPath(chartInstance, selection, clipDimension) {
    let clipId = `sankey-animation-clip-${chartInstance.idNumeric}`,
        clipRect = svg_clips.createElement('rect', clipId, chartInstance);
    clipDimension = clipDimension || { x: 0, y: 0, width: chartInstance.plotarea.width, height: chartInstance.plotarea.height };

    const isSmoothScrollEnabled = Registry.getComponent('scroll.isSmoothScrollEnabled');
    if (isSmoothScrollEnabled(chartInstance)) {
        clipDimension.x = chartInstance.scrolledX || 0;
        clipDimension.y = chartInstance.scrolledY || 0;
        clipDimension.width += chartInstance.scrolledX || 0;
        clipDimension.height += chartInstance.scrolledY || 0;
    }

    // #ZC2587
    clipRect.interrupt();

    svg_clips.setAttributes(clipRect, clipDimension, ['x', 'y', 'width', 'height']);
    selection.attr('clip-path', `url(#${clipId})`);

    let labelElement = chartInstance.container.selectAll('.labelholder');
    labelElement.attr('clip-path', `url(#${clipId})`);

    return clipRect;
}
renderer_helpers.sankey.scrollClipPath = sankeyClipPath;

export function sankeyGradientDimension(point, gradientOptions) {
    if (point && point.exit) {
        gradientOptions.options.linear.y1 = 0;
        gradientOptions.options.linear.y2 = 100;
        gradientOptions.options.linear.x1 = null;
        gradientOptions.options.linear.x2 = null;
    } else {
        gradientOptions.options.linear.x1 = 0;
        gradientOptions.options.linear.x2 = 100;
        gradientOptions.options.linear.y1 = null;
        gradientOptions.options.linear.y2 = null;
    }
}
svg_plot_renderer.sankey = SankeyChart;
canvas_plot_renderer.sankey = SankeyChart;
export var _sankeyChart = SankeyChart;
