// @zohocharts/zc/src/sankey/sankey-circular-tools.js
/// /////////////////////////////////////////////////////////////////////////////////
// Cycle functions

//portion of code to detect circular links based on Colin Fergus' bl.ock https://gist.github.com/cfergus/3956043

// Identify circles in the link objects
import { d3_max, d3_min } from '../core/d3-constants';
import { sankeyCircularLink } from './sankey-circular-link';

/* Given a node, find all links for which this is a source
   in the current 'known' graph  */
function findLinksOutward(node, graph) {
    var children = [];

    for (var i = 0; i < graph.length; i++) {
        if (node == graph[i].source) {
            children.push(graph[i]);
        }
    }

    return children;
}

// Create a normal curve or circular curve
//var curveSankeyForceLink =

// Return the angle between a straight line between the source and target of the link, and the vertical plane of the node
function linkAngle(link) {
    let adjacent = Math.abs(link.y1 - link.y0);
    let opposite = Math.abs(link.target.x0 - link.source.x1);

    return Math.atan(opposite / adjacent);
}

function circularLinksCross(link1, link2) {
    if (link1.source.depth < link2.target.depth) {
        return false;
    } else if (link1.target.depth > link2.source.depth) {
        return false;
    } else {
        return true;
    }
}
export function getCircularLinkMargin(graph, ky, ratio, maxColumn) {
    let verticalMargin = 25;
    let baseRadius = 10;

    var totalTopLinksWidth = 0,
        totalBottomLinksWidth = 0,
        totalRightLinksWidth = 0,
        totalLeftLinksWidth = 0,
        totalLeftSubLinksWidth = {},
        totalRightSubLinksWidth = {},
        linkCount = {
            top: 0,
            bottom: 0,
            left: 0
        };
    /*
    calculate circular link widths
     */
    graph.links.forEach(function(link, i) {
        if (ky != null || ratio != null) {
            if (ky != null) {
                link.width = link.value * ky;
            } else {
                link.width = link.width * ratio;
            }
        }

        if (link.circular) {
            if (link.circularLinkType == 'top') {
                linkCount.top++;
                totalTopLinksWidth = totalTopLinksWidth + link.width;
            } else {
                linkCount.bottom++;
                totalBottomLinksWidth = totalBottomLinksWidth + link.width;
            }

            // if (link.source.column > link.target.column) {
                // totalLeftSubLinksWidth
                //left side extra link width calculation
            
                if (totalLeftSubLinksWidth[link.target.column]) {
                    totalLeftSubLinksWidth[link.target.column].total += link.width;
                } else {
                    totalLeftSubLinksWidth[link.target.column] = {};
                    totalLeftSubLinksWidth[link.target.column].total = link.width;
                }
                
                //ignore the case for link with target column as 0, as we calculate at the end 
                if (link.target.column != 0 && totalLeftSubLinksWidth[link.target.column]) {
                    if (link.target.x0 - totalLeftSubLinksWidth[link.target.column].total < graph.x0) {
                        totalLeftSubLinksWidth[link.target.column].extra =
                            (totalLeftSubLinksWidth[link.target.column].extra || 0) + link.width;
                    }
                }

                //totalRightSubLinksWidth
                //Right side extra link width calculation
                if (totalRightSubLinksWidth[link.source.column]) {
                    totalRightSubLinksWidth[link.source.column].total += link.width;
                } else {
                    totalRightSubLinksWidth[link.source.column] = {};
                    totalRightSubLinksWidth[link.source.column].total = link.width;
                }
                //ignore the case for link with source column as maxColumn, as we calculate at the end 
                if (link.source.column != maxColumn && totalRightSubLinksWidth[link.source.column]) {
                    if (link.source.x1 + totalRightSubLinksWidth[link.source.column].total > graph.x1) {
                        totalRightSubLinksWidth[link.source.column].extra =
                            (totalRightSubLinksWidth[link.source.column].extra || 0) + link.width;
                    }
                }
            // }
            // if (link.source.column < link.target.column) {
            //     //right side link
            //     if (totalRightSubLinksWidth[link.target.column]) {
            //         totalRightSubLinksWidth[link.target.column].total += link.width;
            //     } else {
            //         totalRightSubLinksWidth[link.target.column] = {};
            //         totalRightSubLinksWidth[link.target.column].total = link.width;
            //     }
            //     if (totalRightSubLinksWidth[link.target.column]) {
            //         if (link.target.x1 + totalRightSubLinksWidth[link.target.column].total > graph.x1) {
            //             totalRightSubLinksWidth[link.target.column].extra =
            //                 (totalRightSubLinksWidth[link.target.column].extra || 0) + link.width;
            //         }
            //     }
            // }

            if (link.target.column == 0) {
                totalLeftLinksWidth = totalLeftLinksWidth + link.width;
            }

            if (link.source.column == maxColumn) {
                totalRightLinksWidth = totalRightLinksWidth + link.width;
            }
        }
    });

    let x = Object.keys(totalLeftSubLinksWidth),
        y = Object.keys(totalRightSubLinksWidth);
    (totalLeftLinksWidth +=
        d3_max(x, d => {
            return totalLeftSubLinksWidth[d].extra || 0;
        }) || 0),
        (totalRightLinksWidth +=
            d3_max(y, d => {
                return totalRightSubLinksWidth[d].extra || 0;
            }) || 0);

    //account for radius of curves and padding between links
    totalTopLinksWidth = totalTopLinksWidth > 0 ? totalTopLinksWidth + verticalMargin + baseRadius : totalTopLinksWidth;
    totalBottomLinksWidth = totalBottomLinksWidth > 0 ? totalBottomLinksWidth + verticalMargin + baseRadius : totalBottomLinksWidth;
    totalRightLinksWidth = totalRightLinksWidth > 0 ? totalRightLinksWidth + verticalMargin + baseRadius : totalRightLinksWidth;
    totalLeftLinksWidth = totalLeftLinksWidth > 0 ? totalLeftLinksWidth + verticalMargin + baseRadius : totalLeftLinksWidth;

    /*
   calculate circular link margin
    */
    var circularLinkMargin = {
        top: totalTopLinksWidth,
        bottom: totalBottomLinksWidth + 0,
        left: totalLeftLinksWidth,
        right: totalRightLinksWidth,
        linkCount
    };
    graph.circularSize = {
        width: circularLinkMargin.left + circularLinkMargin.right,
        height: circularLinkMargin.top + circularLinkMargin.bottom,
        hCount: x.length + y.length
    };

    // console.log("hiiiii", circularLinkMargin.top)
    return circularLinkMargin;
}
export function fillHeight(graph, y0, y1, initialExtent) {
    let verticalMargin = 25;
    let baseRadius = 10;
    var nodes = graph.nodes;
    var links = graph.links;

    var top = false;
    var bottom = false;

    links.forEach(function(link) {
        if (link.circularLinkType == 'top') {
            top = true;
        } else if (link.circularLinkType == 'bottom') {
            bottom = true;
        }
    });

    // if (top == false || bottom == false) {
    //     var minY0 = d3_min(nodes, function(node) {
    //         return node.y0;
    //     });
    //     var maxY1 = d3_max(nodes, function(node) {
    //         return node.y1;
    //     });
    //
    //     var minX0 = d3_min(nodes, function(node) {
    //         return node.x0;
    //     });
    //     var maxX1 = d3_max(nodes, function(node) {
    //         return node.x1;
    //     });
    //
    //     var currentHeight = maxY1 - minY0;
    //     var currentWidth = maxX1 - minX0;
    //     var chartHeight = y1 - y0;
    //     var chartWidth = graph.x1 - graph.x0;
    //     var ratio = chartHeight / currentHeight;
    //     var ratio1 = chartWidth / currentWidth;
    //
    //     nodes.forEach(function(node) {
    //         var nodeHeight = (node.y1 - node.y0) * ratio;
    //         node.y0 = (node.y0 - minY0) * ratio;
    //         node.y1 = node.y0 + nodeHeight;
    //
    //         // var nodeWidth = (node.x1 - node.x0) * ratio1;
    //         // node.x0 = (node.x0 - minX0) * ratio1;
    //         // node.x1 = node.x0 + nodeWidth;
    //     });
    //
    //     links.forEach(function(link) {
    //         link.y0 = (link.y0 - minY0) * ratio;
    //         link.y1 = (link.y1 - minY0) * ratio;
    //         link.width = link.width * ratio;
    //     });
    // } else {
    // return;
    var totalTopLinksWidth = 0,
        totalBottomLinksWidth = 0,
        totalRightLinksWidth = 0,
        totalLeftLinksWidth = 0;

    var maxColumn = d3_max(graph.nodes, function(node) {
        return node.column;
    });
    var m = 0;
    let margin = getCircularLinkMargin(graph, null, null, maxColumn);
    // graph.links.forEach(function(link) {
    //     if (link.circular) {
    //         if (link.circularLinkType == 'top') {
    //             m = Math.max(link.totalHeight, m);
    //             totalTopLinksWidth = totalTopLinksWidth + link.width;
    //         } else {
    //             totalBottomLinksWidth = totalBottomLinksWidth + link.width;
    //         }
    //
    //         if (link.target.column == 0) {
    //             totalLeftLinksWidth = totalLeftLinksWidth + link.width;
    //         }
    //
    //         if (link.source.column == maxColumn) {
    //             totalRightLinksWidth = totalRightLinksWidth + link.width;
    //         }
    //     }
    // });
    //
    // //account for radius of curves and padding between links
    // totalTopLinksWidth = totalTopLinksWidth > 0 ? totalTopLinksWidth + verticalMargin + baseRadius : totalTopLinksWidth;
    // totalBottomLinksWidth = totalBottomLinksWidth > 0 ? totalBottomLinksWidth + verticalMargin + baseRadius : totalBottomLinksWidth;
    // totalRightLinksWidth = totalRightLinksWidth > 0 ? totalRightLinksWidth + verticalMargin + baseRadius : totalRightLinksWidth;
    // totalLeftLinksWidth = totalLeftLinksWidth > 0 ? totalLeftLinksWidth + verticalMargin + baseRadius : totalLeftLinksWidth;

    // var margin = {
    //     top: totalTopLinksWidth,
    //     bottom: totalBottomLinksWidth,
    //     left: totalLeftLinksWidth,
    //     right: totalRightLinksWidth
    // };

    var maxColumn = d3_max(graph.nodes, function(node) {
        return node.column;
    });
    var minY0 = d3_min(nodes, function(node) {
        return node.y0;
    });
    var maxY1 = d3_max(nodes, function(node) {
        return node.y1;
    });

    // var currentWidth = graph.x1 - graph.x0;
    // var currentHeight = graph.y1 - graph.y0;

    // var newWidth = currentWidth + margin.right + margin.left;
    let bottomLinkgap = margin.linkCount.bottom * 1;
    var newHeight = (y1 = graph.y1 = initialExtent.y1 - margin.bottom - bottomLinkgap);
    return;
    // var ratio = newHeight / maxY1;
    var ratio = newHeight / maxY1;
    var dy = minY0 > margin.top ? minY0 - margin.top - 5 : 0;
    nodes.forEach(function(node) {
        var nodeHeight = (node.y1 - node.y0) * ratio;
        node.y0 = (node.y0 - dy) * ratio;
        node.y1 = node.y0 + nodeHeight;
    });

    links.forEach(function(link) {
        link.y0 = (link.y0 - dy) * ratio;
        link.y1 = (link.y1 - dy) * ratio;
        link.width = link.width * ratio;
    });

    // var newHeight1 = (y0 = graph.y0 = initialExtent.y0 + margin.top);
    // var dy = minY0 > margin.top ? minY0 - margin.top : 0;
    //
    // var ratio1 = newHeight1/minY0;
    //
    // nodes.forEach(function(node) {
    //     var nodeHeight = (node.y1 - node.y0) * ratio1;
    //     node.y0 = (node.y0 - 0) * ratio1;
    //     node.y1 = node.y0 + nodeHeight;
    // });
    //
    // links.forEach(function(link) {
    //     link.y0 = (link.y0 - 0) * ratio1;
    //     link.y1 = (link.y1 - 0) * ratio1;
    //     link.width = link.width * ratio1;
    // });
    // }
}
function calcVerticalBuffer(links) {
    let circularLinkGap = 2;
    links.sort(sortLinkDepthAscending);
    links.forEach(function(link, i) {
        let buffer = 0;

        let j = 0;
        for (j; j < i; j++) {
            if (circularLinksCross(links[i], links[j])) {
                let bufferOverThisLink = links[j].circularPathData.verticalBuffer1 + links[j].width / 2 + circularLinkGap;
                buffer = bufferOverThisLink > buffer ? bufferOverThisLink : buffer;
            }
        }
        link.circularPathData.verticalBuffer1 = buffer + link.width / 2;
        link.circularPathData.verticalBuffer = buffer + link.width;
    });

    return links;
}
// sort links based on the distance between the source and tartget node depths
// if the same, then use Y position of the source node
function sortLinkDepthAscending(link1, link2) {
    if (linkDepthDistance(link1) == linkDepthDistance(link2)) {
        return link1.circularLinkType == 'bottom' ? sortLinkSourceYDescending(link1, link2) : sortLinkSourceYAscending(link1, link2);
    } else {
        // return linkDepthDistance(link1) - linkDepthDistance(link2);
        return linkDepthDistance(link2) - linkDepthDistance(link1);
    }
}

function sortLinkSourceYAscending(link1, link2) {
    return link1.y0 - link2.y0;
}

function sortLinkSourceYDescending(link1, link2) {
    return link2.y0 - link1.y0;
}

function sortLinkTargetYAscending(link1, link2) {
    return link1.y1 - link2.y1;
}

function sortLinkTargetYDescending(link1, link2) {
    return link2.y1 - link1.y1;
}

// return the distance between the link's target and source node, in terms of the nodes' depth
function linkDepthDistance(link) {
    return link.target.depth - link.source.depth;
}

// return the distance between the link's target and source node, in terms of the nodes' X coordinate
function linkXLength(link) {
    return link.target.x0 - link.source.x1;
}

function linkPerpendicularYToLinkSource(longerLink, shorterLink) {
    // Return the Y coordinate on the longerLink path * which is perpendicular shorterLink's source.
    // * approx, based on a straight line from target to source, when in fact the path is a bezier

    // get the angle for the longer link
    let angle = linkAngle(longerLink);

    // get the adjacent length to the other link's x position
    let heightFromY1ToPependicular = linkXLength(shorterLink) / Math.tan(angle);

    // add or subtract from longer link1's original y1, depending on the slope
    let yPerpendicular =
        incline(longerLink) == 'up' ? longerLink.y1 + heightFromY1ToPependicular : longerLink.y1 - heightFromY1ToPependicular;

    return yPerpendicular;
}

function linkPerpendicularYToLinkTarget(longerLink, shorterLink) {
    // Return the Y coordinate on the longerLink path * which is perpendicular shorterLink's source.
    // * approx, based on a straight line from target to source, when in fact the path is a bezier

    // get the angle for the longer link
    let angle = linkAngle(longerLink);

    // get the adjacent length to the other link's x position
    let heightFromY1ToPependicular = linkXLength(shorterLink) / Math.tan(angle);

    // add or subtract from longer link's original y1, depending on the slope
    let yPerpendicular =
        incline(longerLink) == 'up' ? longerLink.y1 - heightFromY1ToPependicular : longerLink.y1 + heightFromY1ToPependicular;

    return yPerpendicular;
}

export function resolveNodeLinkOverlaps(graph) {
    var minY0 = d3_min(graph.nodes, function(node) {
        return node.y0;
    });
    var maxY1 = d3_max(graph.nodes, function(node) {
        return node.y1;
    });
    var extent = { y0: minY0, y1: maxY1 };
    graph.links.forEach(function(link) {
        if (link.circular) {
            return;
        }

        if (link.target.depth - link.source.depth > 1) {
            let depthToTest = link.source.depth + 1;
            let maxDepthToTest = link.target.depth - 1;

            let i = 1;
            let numberOfDepthsToTest = maxDepthToTest - depthToTest + 1;

            for (depthToTest, i = 1; depthToTest <= maxDepthToTest; depthToTest++, i++) {
                adjustNodePosition(link, graph, extent, depthToTest, numberOfDepthsToTest, i);
            }
        }
    });
}

function adjustNodePosition(link, graph, extent, depthToTest, numberOfDepthsToTest, i) {
    graph.nodes.forEach(node => {
        if (node.depth == depthToTest) {
            let t = i / (numberOfDepthsToTest + 1);

            // Find all the points of a cubic bezier curve in javascript
            // https://stackoverflow.com/questions/15397596/find-all-the-points-of-a-cubic-bezier-curve-in-javascript

            let B0_t = Math.pow(1 - t, 3);
            let B1_t = 3 * t * Math.pow(1 - t, 2);
            let B2_t = 3 * Math.pow(t, 2) * (1 - t);
            let B3_t = Math.pow(t, 3);

            let py_t = B0_t * link.y0 + B1_t * link.y0 + B2_t * link.y1 + B3_t * link.y1;

            let linkY0AtDepth = py_t - link.width / 2;
            let linkY1AtDepth = py_t + link.width / 2;

            // If top of link overlaps node, push node up
            if (linkY0AtDepth > node.y0 && linkY0AtDepth < node.y1) {
                // console.log("OVERLAP!")

                let dy = -(node.y1 - linkY0AtDepth + 10);

                //check if other nodes need to move up too
                var nodeListToCheck = getOverlappedNodes(graph, node);

                let leavesOut = checkAnyNodeLeavesOut(nodeListToCheck, extent, dy);
                if (!leavesOut) {
                    nodeListToCheck.map(node => {
                        adjustNodeHeight(node, dy);
                    });
                }
            } else if (linkY1AtDepth > node.y0 && linkY1AtDepth < node.y1) {
                // If bottom of link overlaps node, push node down
                let dy = linkY1AtDepth - node.y0 + 10;

                //check if other nodes need to move down too
                var nodeListToCheck = getOverlappedNodes(graph, node, true);

                let leavesOut = checkAnyNodeLeavesOut(nodeListToCheck, extent, dy);
                if (!leavesOut) {
                    nodeListToCheck.map(node => {
                        adjustNodeHeight(node, dy);
                    });
                }
            } else if (linkY0AtDepth < node.y0 && linkY1AtDepth > node.y1) {
                // if link completely overlaps node
                let dy = linkY1AtDepth - node.y0 + 10;

                var nodeListToCheck = getOverlappedNodes(graph, node, true);

                let leavesOut = checkAnyNodeLeavesOut(nodeListToCheck, extent, dy);
                if (!leavesOut) {
                    nodeListToCheck.map(node => {
                        adjustNodeHeight(node, dy);
                    });
                }
            }
        }
    });
}

function getOverlappedNodes(graph, node, checkOnlyBottom) {
    let arr = [node];
    //check if other nodes need to move up too
    graph.nodes.forEach(function(otherNode) {
        //don't need to check itself or nodes at different depths
        if (otherNode.name == node.name || otherNode.depth != node.depth) {
            return;
        }
        if (checkOnlyBottom) {
            if (otherNode.y0 < node.y1 && otherNode.y1 > node.y1) {
                arr.push(otherNode);
            }
        } else {
            if (nodesOverlap(node, otherNode)) {
                arr.push(otherNode);
            }
        }
    });
    return arr;
}
function nodesOverlap(nodeA, nodeB) {
    //test if nodeA top partially overlaps nodeB
    if (nodeA.y0 > nodeB.y0 && nodeA.y0 < nodeB.y1) {
        return true;
    }
    //test if nodeA bottom partially overlaps nodeB
    else if (nodeA.y1 > nodeB.y0 && nodeA.y1 < nodeB.y1) {
        return true;
    }
    //test if nodeA covers nodeB
    else if (nodeA.y0 < nodeB.y0 && nodeA.y1 > nodeB.y1) {
        return true;
    } else {
        return false;
    }
}

function adjustNodeHeight(node, dy) {
    node.y0 = node.y0 + dy;
    node.y1 = node.y1 + dy;

    node.targetLinks.forEach(function(l) {
        l.y1 = l.y1 + dy;
    });

    node.sourceLinks.forEach(function(l) {
        l.y0 = l.y0 + dy;
    });

    return node;
}
function checkAnyNodeLeavesOut(nodes, extent, dy) {
    let leavesOutside = nodes.some(node => {
        //isWithInExtent
        return !(node.y0 + dy > extent.y0 && node.y1 + dy < extent.y1);
    });
    return leavesOutside;
}
export function sortSourceLinks(graph, height) {
    graph.nodes.forEach(function(node) {
        // move any nodes up which are off the bottom
        if (node.y + (node.y1 - node.y0) > height) {
            node.y = node.y - (node.y + (node.y1 - node.y0) - height);
        }

        let nodesSourceLinks = graph.links.filter(function(l) {
            return l.source.name == node.name;
        });

        // if more than 1 link then sort
        if (nodesSourceLinks.length > 1) {
            nodesSourceLinks.sort(function(link1, link2) {
                // if both are not circular...
                if (!link1.circular && !link2.circular) {
                    // if the target nodes are the same depth, then sort by the link's target y
                    if (link1.target.depth == link2.target.depth) {
                        return link1.y1 - link2.y1;
                    } else if (!sameInclines(link1, link2)) {
                        // if the links slope in different directions, then sort by the link's target y
                        return link1.y1 - link2.y1;

                        // if the links slope in same directions, then sort by any overlap
                    } else {
                        if (link1.target.depth > link2.target.depth) {
                            // if (node.name == "process10") {console.log("here")}
                            /* let link2Angle = linkAngleFromSource(link2);
                                let link2AdjToLink1Y = linkXLength(link1) / Math.tan(link2Angle);
                                let link2Adj = incline(link2) == "up"
                                  ? link2.y0 - link2AdjToLink1Y
                                  : link2.y0 + link2AdjToLink1Y; */

                            let link2Adj = linkPerpendicularYToLinkTarget(link2, link1);

                            return link1.y1 - link2Adj;
                        }
                        if (link2.target.depth > link1.target.depth) {
                            /* let link1Angle = linkAngleFromSource(link1);
                                let link1AdjToLink2Y = linkXLength(link2) / Math.tan(link1Angle);
                                let link1Adj = incline(link1) == "up"
                                  ? link1.y0 - link1AdjToLink2Y
                                  : link1.y0 + link1AdjToLink2Y; */

                            let link1Adj = linkPerpendicularYToLinkTarget(link1, link2);

                            return link1Adj - link2.y1;
                        }
                    }
                }

                // if only one is circular, the move top links up, or bottom links down
                if (link1.circular && !link2.circular) {
                    return link1.circularLinkType == 'top' ? -1 : 1;
                } else if (link2.circular && !link1.circular) {
                    return link2.circularLinkType == 'top' ? 1 : -1;
                }

                // if both links are circular...
                if (link1.circular && link2.circular) {
                    // ...and they both loop the same way (both top)
                    if (link1.circularLinkType === link2.circularLinkType && link1.circularLinkType == 'top') {
                        // ...and they both connect to a target with same depth, then sort by the target's y
                        if (link1.target.depth === link2.target.depth) {
                            return link1.target.y1 - link2.target.y1;
                        } else {
                            // ...and they connect to different depth targets, then sort by how far back they
                            return link2.target.depth - link1.target.depth;
                        }
                    } else if (link1.circularLinkType === link2.circularLinkType && link1.circularLinkType == 'bottom') {
                        // ...and they both loop the same way (both bottom)
                        // ...and they both connect to a target with same depth, then sort by the target's y
                        if (link1.target.depth === link2.target.depth) {
                            return link2.target.y1 - link1.target.y1;
                        } else {
                            // ...and they connect to different depth targets, then sort by how far back they
                            return link1.target.depth - link2.target.depth;
                        }
                    } else {
                        // ...and they loop around different ways, the move top up and bottom down
                        return link1.circularLinkType == 'top' ? -1 : 1;
                    }
                }
            });
        }

        // update y0 for links
        let ySourceOffset = node.y0;

        nodesSourceLinks.forEach(function(link) {
            link.y0 = ySourceOffset + link.width / 2;
            ySourceOffset = ySourceOffset + link.width;
        });
    });
}

export function sortTargetLinks(graph, height) {
    graph.nodes.forEach(function(node) {
        let nodesTargetLinks = graph.links.filter(function(l) {
            return l.target.name == node.name;
        });

        if (nodesTargetLinks.length > 1) {
            nodesTargetLinks.sort(function(link1, link2) {
                // if both are not circular, the base on the source y position
                if (!link1.circular && !link2.circular) {
                    if (link1.source.depth == link2.source.depth) {
                        return link1.y0 - link2.y0;
                    } else if (!sameInclines(link1, link2)) {
                        return link1.y0 - link2.y0;
                    } else {
                        // get the angle of the link to the further source node (ie the smaller depth)
                        if (link2.source.depth < link1.source.depth) {
                            let link2Adj = linkPerpendicularYToLinkSource(link2, link1);

                            return link1.y0 - link2Adj;
                        }
                        if (link1.source.depth < link2.source.depth) {
                            let link1Adj = linkPerpendicularYToLinkSource(link1, link2);

                            return link1Adj - link2.y0;
                        }
                    }
                }

                // if only one is circular, the move top links up, or bottom links down
                if (link1.circular && !link2.circular) {
                    return link1.circularLinkType == 'top' ? -1 : 1;
                } else if (link2.circular && !link1.circular) {
                    return link2.circularLinkType == 'top' ? 1 : -1;
                }

                // if both links are circular...
                if (link1.circular && link2.circular) {
                    // ...and they both loop the same way (both top)
                    if (link1.circularLinkType === link2.circularLinkType && link1.circularLinkType == 'top') {
                        // ...and they both connect to a target with same depth, then sort by the target's y
                        if (link1.source.depth === link2.source.depth) {
                            return link1.source.y1 - link2.source.y1;
                        } else {
                            // ...and they connect to different depth targets, then sort by how far back they
                            return link1.source.depth - link2.source.depth;
                        }
                    } else if (link1.circularLinkType === link2.circularLinkType && link1.circularLinkType == 'bottom') {
                        // ...and they both loop the same way (both bottom)
                        // ...and they both connect to a target with same depth, then sort by the target's y
                        if (link1.source.depth === link2.source.depth) {
                            return link1.source.y1 - link2.source.y1;
                        } else {
                            // ...and they connect to different depth targets, then sort by how far back they
                            return link2.source.depth - link1.source.depth;
                        }
                    } else {
                        // ...and they loop around different ways, the move top up and bottom down
                        return link1.circularLinkType == 'top' ? -1 : 1;
                    }
                }
            });
        }

        // update y1 for links
        let yTargetOffset = node.y0;

        nodesTargetLinks.forEach(function(link) {
            link.y1 = yTargetOffset + link.width / 2;
            yTargetOffset = yTargetOffset + link.width;
        });
    });
}

function sameInclines(link1, link2) {
    return incline(link1) == incline(link2);
}

function incline(link) {
    // positive => slopes up from source to target
    // negative => slopes down from source to target
    return link.y0 - link.y1 > 0 ? 'up' : 'down';
}

export function addCircularPathData(graph, linkPadding) {
    let baseRadius = 10;
    let buffer = 10;
    let verticalMargin = 25;
    let circularLinkGap = 2,
        height = graph.y1;

    let minY = d3_min(graph.links, function(link) {
        return link.source.y0;
    });

    // create object for circular Path Data
    graph.links.forEach(function(link) {
        if (link.circular) {
            link.circularPathData = {};
        }
    });

    // calc vertical offsets per top/bottom links
    let topLinks = graph.links.filter(function(l) {
        return l.circularLinkType == 'top';
    });
    topLinks = calcVerticalBuffer(topLinks);

    let bottomLinks = graph.links.filter(function(l) {
        return l.circularLinkType == 'bottom';
    });
    bottomLinks = calcVerticalBuffer(bottomLinks);

    // add the base data for each link
    graph.links.forEach(function(link) {
        if (link.circular) {
            link.circularPathData.arcRadius = link.width + baseRadius;
            link.circularPathData.leftNodeBuffer = buffer;
            link.circularPathData.rightNodeBuffer = buffer;
            link.circularPathData.sourceWidth = link.source.x1 - link.source.x0;

            if (link.circularLinkType == 'bottom') {
                link.circularPathData.sourceX = link.source.x0 + link.circularPathData.sourceWidth + linkPadding;
                link.circularPathData.targetX = link.target.x0 - linkPadding;
                link.circularPathData.sourceY = link.y0 - link.width / 2;
                link.circularPathData.targetY = link.y1 - link.width / 2;
            } else {
                link.circularPathData.sourceX = link.source.x0 + link.circularPathData.sourceWidth + linkPadding;
                link.circularPathData.targetX = link.target.x0 - linkPadding;
                link.circularPathData.sourceY = link.y0 + link.width / 2;
                link.circularPathData.targetY = link.y1 + link.width / 2;
            }

            // add left extent coordinates, based on links with same source depth and circularLink type
            let thisDepth = link.source.depth;
            let thisCircularLinkType = link.circularLinkType;
            let sameDepthLinks = graph.links.filter(function(l) {
                return l.source.depth == thisDepth && l.circularLinkType == thisCircularLinkType;
            });

            if (link.circularLinkType == 'bottom') {
                sameDepthLinks.sort(sortLinkSourceYDescending);
            } else {
                sameDepthLinks.sort(sortLinkSourceYAscending);
            }

            let radiusOffset = 0;
            sameDepthLinks.forEach(function(l, i) {
                if (l.circularLinkID == link.circularLinkID) {
                    link.circularPathData.leftSmallArcRadius = baseRadius + link.width + radiusOffset;
                    link.circularPathData.leftLargeArcRadius = baseRadius + link.width + i * circularLinkGap + radiusOffset;

                    link.circularPathData.leftSmallArcRadius1 = link.circularPathData.leftSmallArcRadius - link.width;
                    link.circularPathData.leftLargeArcRadius1 = link.circularPathData.leftLargeArcRadius - link.width;
                }
                radiusOffset = radiusOffset + l.width;
            });

            // add right extent coordinates, based on links with same target depth and circularLink type
            thisDepth = link.target.depth;
            sameDepthLinks = graph.links.filter(function(l) {
                return l.target.depth == thisDepth && l.circularLinkType == thisCircularLinkType;
            });
            if (link.circularLinkType == 'bottom') {
                sameDepthLinks.sort(sortLinkTargetYDescending);
            } else {
                sameDepthLinks.sort(sortLinkTargetYAscending);
            }

            radiusOffset = 0;
            sameDepthLinks.forEach(function(l, i) {
                if (l.circularLinkID == link.circularLinkID) {
                    link.circularPathData.rightSmallArcRadius = baseRadius + link.width + radiusOffset;
                    link.circularPathData.rightLargeArcRadius = baseRadius + link.width + i * circularLinkGap + radiusOffset;

                    link.circularPathData.rightSmallArcRadius1 = link.circularPathData.rightSmallArcRadius - link.width;
                    link.circularPathData.rightLargeArcRadius1 = link.circularPathData.rightLargeArcRadius - link.width;
                }
                radiusOffset = radiusOffset + l.width;
            });

            // all links
            link.circularPathData.leftInnerExtent = link.circularPathData.sourceX + link.circularPathData.leftNodeBuffer;
            link.circularPathData.rightInnerExtent = link.circularPathData.targetX - link.circularPathData.rightNodeBuffer;
            link.circularPathData.leftFullExtent =
                link.circularPathData.sourceX + link.circularPathData.leftLargeArcRadius + link.circularPathData.leftNodeBuffer;
            link.circularPathData.rightFullExtent =
                link.circularPathData.targetX - link.circularPathData.rightLargeArcRadius - link.circularPathData.rightNodeBuffer;

            link.circularPathData.rightFullExtent1 = link.circularPathData.rightFullExtent + link.width;
            link.circularPathData.rightInnerExtent1 = link.circularPathData.rightFullExtent1 + link.circularPathData.rightLargeArcRadius1;

            link.circularPathData.leftFullExtent1 = link.circularPathData.leftFullExtent - link.width;
            link.circularPathData.leftInnerExtent1 = link.circularPathData.leftFullExtent1 - link.circularPathData.leftLargeArcRadius1;

            link.circularPathData.leftWidth = link.circularPathData.leftLargeArcRadius + link.circularPathData.leftNodeBuffer;
            link.circularPathData.rightWidth = link.circularPathData.rightLargeArcRadius + link.circularPathData.rightNodeBuffer;
            // bottom links
            if (link.circularLinkType == 'bottom') {
                link.circularPathData.verticalFullExtent = height + verticalMargin + link.circularPathData.verticalBuffer;
                link.circularPathData.verticalLeftInnerExtent =
                    link.circularPathData.verticalFullExtent - link.circularPathData.leftLargeArcRadius;
                link.circularPathData.verticalRightInnerExtent =
                    link.circularPathData.verticalFullExtent - link.circularPathData.rightLargeArcRadius;

                link.circularPathData.verticalFullExtent1 = height + verticalMargin + link.circularPathData.verticalBuffer - link.width;
                link.circularPathData.verticalRightInnerExtent1 = link.circularPathData.verticalFullExtent1;

                link.circularPathData.verticalLeftInnerExtent1 = link.circularPathData.verticalFullExtent1;
                link.circularPathData.totalHeight = verticalMargin + link.circularPathData.verticalBuffer;
                // link.circularPathData.verticalRightInnerExtent1 = link.circularPathData.verticalFullExtent - link.circularPathData.rightLargeArcRadius/2
            } else {
                // top links
                link.circularPathData.verticalFullExtent = minY - verticalMargin - link.circularPathData.verticalBuffer;
                link.circularPathData.verticalLeftInnerExtent =
                    link.circularPathData.verticalFullExtent + link.circularPathData.leftLargeArcRadius;
                link.circularPathData.verticalRightInnerExtent =
                    link.circularPathData.verticalFullExtent + link.circularPathData.rightLargeArcRadius;

                link.circularPathData.verticalFullExtent1 = link.circularPathData.verticalFullExtent + link.width;
                link.circularPathData.verticalLeftInnerExtent1 = link.circularPathData.verticalFullExtent1;
                link.circularPathData.verticalRightInnerExtent1 = link.circularPathData.verticalFullExtent1;

                link.circularPathData.totalHeight = verticalMargin + link.circularPathData.verticalBuffer;
            }

            link.circularPathData.path = sankeyCircularLink(link);
        }
    });
}
