// @zohocharts/zc/src/legend/title/legend-title-svg-parser.js
//
import { TitleParserSvg } from '../../title/title-parser';
import { Mark } from '../../mark/mark-utils';
import { pick, parseShortHandValue, setPropVal } from '../../utils/zcutils';

function LegendTitleSvgParser(spec) {
    let { config, data = [1] } = spec,
        { show } = config,
        margin = parseShortHandValue(pick(config.margin, 3));

    // Wrap
    let groupMark = Mark('group', '.zc-legend-title', data);

    if (show) {
        let titleMark = TitleParserSvg(spec);
        titleMark.as = { x: 'bb_x', y: 'bb_y', width: 'bb_width', height: 'bb_height' };

        let backgroundMark = Mark('_rect', null, data);
        backgroundMark.config = {
            x: asParser('bb_x', -margin.left),
            y: asParser('bb_y', -margin.top),
            width: asParser('bb_width', margin.left + margin.right),
            height: asParser('bb_height', margin.top + margin.bottom)
        };
        setPropVal(backgroundMark, 'properties.styles', {
            opacity: 0,
            display: () => {
                let width = titleMark._as[titleMark.as.width][0],
                    height = titleMark._as[titleMark.as.height][0];
                return !(width || height) ? 'none' : null;
            }
        });
        groupMark.children = [titleMark, backgroundMark];
    }

    return groupMark;
}

function asParser(as, margin) {
    return { _type: 'as', as, onEnd: _ => _ + margin };
}

export { LegendTitleSvgParser };
