// @zohocharts/zc/src/legend/title/legend-title-html-parser.js
//
import { VERTICAL_ALIGN } from '../../core/align-constants';
import { ListParser } from '../../view/list/list-parser';
import { decodeHTML, encodeHTML, getValByPriority, mergeJSON, utils } from '../../utils/zcutils';
import { $ZCG } from '../../core/zcglobal';
import { isFunction } from '../../utils/function-utils';

function LegendTitleHtmlParser(spec) {
    let chartObj = this.chartObj;
    let { config, globalFont, computedInfo, onEnd } = spec;

    let titleConfig = config.title || {},
        boxConfig = mergeJSON(config.list.marker, config.selectAllBox, true),
        labelConfig = { ...titleConfig, keepAligned: true, padding: config.textPadding },
        c = {
            show: titleConfig.show,
            itemMargin: titleConfig.margin,
            list: { marker: { ...boxConfig, show: config.selectAllBox && config.selectAllBox.show ? true : false}, label: {...labelConfig} }
        },
        isSizeLegend = spec.encode === "z";

    let viewSpec = {
        config: c,
        data: [titleConfig.text],
        align: { layout: VERTICAL_ALIGN },
        format: d => isFunction(d) ? d(chartObj) : encodeHTML(d),
        globalFont: globalFont,
        label: {
            //isSizeLegend condition is included inorder to avoid image diff in size legend, in future both sizelegend and discrete legend should work with common code //TODO
            fontSize: +utils.getVal(isSizeLegend ? titleConfig.fontSize : config.fontSize || globalFont.fontSize)
        },
        class: 'zc-header-', //remove
        computedInfo,
        onEnd: onEnd,
        selectAllConf: config.selectAllBox
    };

    return ListParser(viewSpec);
}

export { LegendTitleHtmlParser };
