// @zohocharts/zc/src/components/marker/marker-behaviour-mapping.js
//
import { highlighter_c } from '../../events/visualEffects/controllers/highlighter_c';
import { highlightTouch_c } from '../../events/visualEffects/controllers/highlight-touch_c';
import { hidetooltip_c } from '../../events/visualEffects/controllers/hidetooltip_c';
import { reset_c } from '../../events/visualEffects/controllers/reset_c';
import { showCursor_C } from '../../events/visualEffects/controllers/cursor_c';
import { resethighlight_c } from '../../events/visualEffects/controllers/resethighlight_c';
import { conditionalDefault } from '../../events/visualEffects/behaviour-mapping';

let markerBehaviourMap = {
    mousemove: [
        {
            action: () => {
                return resethighlight_c;
            },
            allowifcustom: true,
            type: 'resetHighlight'
        },
        {
            action: () => {
                return highlighter_c;
            },
            type: 'highlight'
        },
        {
            action: () => {
                return showCursor_C;
            },
            type: 'cursor'
        },
        {
            action: {
                callback: function(event, d, chart) {
                    var tooltip = chart.eventHandler.tooltipElement;
                    tooltip.hide(tooltip.element);
                }
            },
            type: 'custom'
        }
    ],
    click: [
        {
            action: () => {
                return hidetooltip_c;
            },
            type: 'hideTooltip'
        },
        {
            action: () => {
                return showCursor_C;
            },
            type: 'cursor',
            allowifcustom: conditionalDefault
        },
        {
            action: () => {
                return highlightTouch_c;
            },
            type: 'highlightOnTaphold'
        }
    ],
    doubleclick: [
        {
            action: () => {
                return hidetooltip_c;
            },
            type: 'hideTooltip'
        },
        {
            action: () => {
                return showCursor_C;
            },
            type: 'cursor',
            allowDefault: conditionalDefault
        }
    ],
    mouseout: [
        {
            action: () => {
                return resethighlight_c;
            },
            type: 'resetHighlight',
            allowDefault: true
        },
        {
            action: () => {
                return reset_c;
            },
            type: 'resetTooltip',
            allowDefault: true
        }
    ],
    contextmenu: []
};

markerBehaviourMap.touchmove = markerBehaviourMap.mousemove;
markerBehaviourMap.tap = markerBehaviourMap.click;
markerBehaviourMap.doubletap = markerBehaviourMap.doubleclick;

export { markerBehaviourMap };
