// @zohocharts/zc/src/components/marker/marker-highlight.js
//
import { ScatterHighlighter } from '../../scatter/scatterHighlight';
import { Registry } from '../../core/zc-registry';
import { appendEle } from '../../utils/zcsvgutils';
import { addSVGmarker, getSegmentsForSymbol } from '../../utils/svg-symbols';
import { inherit } from '../../utils/zcutils';
import { d3_select } from '../../core/d3-constants';
import { MarkerParser } from '../../utils/marker-parser';
import { renderDom } from '../../mark/dom-render';

export function Markerhighlighter(chart, helpers) {
    this.chart = chart;
    this.helpers = helpers;
}

inherit(Markerhighlighter, ScatterHighlighter);

let proto = Markerhighlighter.prototype;
proto.options = function(optionInfo) {
    this.optionsInfo = optionInfo;
};
proto.data = function() {
    let chart = this.chart,
        helpers = this.helpers;
    return function(p) {
        let point = this;
        if (!chart.cache.data[point._key]) {
            chart.cache.data[point._key] = {};
        }
        var cache = chart.cache.data[point._key];
        cache._x = point._x = helpers.x(point);
        cache._y = point._y = helpers.y(point);
        cache._size = point._size = helpers.options.size;
        cache._visible = point._visible = (point._dependents && point._dependents.to && point._dependents.to._x) ? point._dependents.to._x > 0 : true;
        return point;
    };
};
proto.getSegments = function(point) {
    let data = point._data(),
        symbol = this.helpers.highlightOptions.segment || this.helpers.options.shape,
        segment = getSegmentsForSymbol(symbol, data._size, data._x, data._y);
    return [segment];
};
proto.path = function(point, model) {
    point = (point._data && point._data()) || point;
    var _options = this.optionsInfo;
    var context = model.context.call(this),
        translate = model.translate.call(this),
        { seriesIndex, paneid } = point;
    _options.options.strokeWidth = 2;

    context.style('display', '');
    context.attr('transform', `translate(${translate[0]},${translate[1]})`);

    let arrowMark = MarkerParser({
            selection: '.outer',
            data: [point],
            config: {
                x: point._x,
                y: point._y,
                ..._options.options
            }
        }),
        _conversionLineMark = {
            mark: {
                mark: 'g',
                selector: `.highlight${point.paneid}${seriesIndex}${point.itemIndex}${model.type}` || '_',
                data: [point],
                properties: {
                    attrs: {
                        display: ''
                    }
                },
                children: [
                    {
                        ...arrowMark
                    }
                ]
            }
        };
    renderDom(_conversionLineMark.mark, context);
};

Registry.setComponent('events.markerhighlighter', Markerhighlighter);
