// @zohocharts/zc/src/bar/parser/radial/marks/radialbar-marks.js
import { ParserRegistry } from "../../../../core/parser-registry";
import { deleteProperties, isArray, isNumber, mergeJSON, perToPx } from "../../../../utils/zcutils";
import { _mainBar as bar } from "../../../bar-marks";

const STARTANGLE = 'startAngle',
    ENDANGLE = 'endAngle',
    INNERRADIUS = 'innerRadius',
    OUTERRADIUS = 'outerRadius',
    CORNERRADIUS = 'cornerRadius',
    PADANGLE = 'padAngle',
    ENTER = 'enter',
    UPDATE = 'update',
    EXIT = 'exit';

export function _mainRadialBar(data) {
    const KEY = 'radialbar';
    const self = this;

    const { chart, config, helpers, builderC } = this,
        anime_config = builderC.animationOptions;

    const { position, styles } = helpers,
        _position = position[KEY];

    const animeFunction = function (value = STARTANGLE, type = UPDATE) {
        return function () { return _position[value][type].apply(this, arguments) };
    }

    const parser = bar.call(this, data);
    const custom = {
        type: 'HTML',
        children: [{
            mark: '_arc',
            data: function (d, i) {
                return d.disabled ? [] : self.helpers.filterData(d)
            },
            // dataFilter: function (d, i) {
            //     return d.disabled ? null : d
            // },
            anime: {
                enter: {
                    [STARTANGLE]: animeFunction(STARTANGLE, ENTER),
                    [ENDANGLE]: animeFunction(ENDANGLE, ENTER),
                    [INNERRADIUS]: animeFunction(INNERRADIUS, ENTER),
                    [OUTERRADIUS]: animeFunction(OUTERRADIUS, ENTER),
                    [CORNERRADIUS]: animeFunction(CORNERRADIUS, ENTER),
                    [PADANGLE]: animeFunction(PADANGLE, ENTER)
                },
                update: {
                    [STARTANGLE]: animeFunction(STARTANGLE),
                    [ENDANGLE]: animeFunction(ENDANGLE),
                    [INNERRADIUS]: animeFunction(INNERRADIUS),
                    [OUTERRADIUS]: animeFunction(OUTERRADIUS),
                    [CORNERRADIUS]: animeFunction(CORNERRADIUS),
                    [PADANGLE]: animeFunction(PADANGLE),
                    duration: anime_config._duration,
                    delay: anime_config._delay,
                    easing: anime_config.easingType
                },
                exit: {
                    [STARTANGLE]: animeFunction(STARTANGLE, EXIT),
                    [ENDANGLE]: animeFunction(ENDANGLE, EXIT),
                    [INNERRADIUS]: animeFunction(INNERRADIUS, EXIT),
                    [OUTERRADIUS]: animeFunction(OUTERRADIUS, EXIT),
                    [CORNERRADIUS]: animeFunction(CORNERRADIUS, EXIT),
                    [PADANGLE]: animeFunction(PADANGLE, EXIT),
                    duration: anime_config._duration
                }
            },
            properties: {
                attrs: {
                    class: (d, i) => `zc_bar zc_bar_${d.itemIndex || i}`
                }
            }
        }]
    }
    deleteProperties(parser.children[0], ['config', 'anime'])
    mergeJSON(parser, custom, true, true);
    return parser;
}
ParserRegistry.marks('radialbar.default', _mainRadialBar);
