// @zohocharts/zc/src/button/custom-button.js
import { createElement, setStyles } from '../utils/zcsvgutils';
import { getMargin } from '../axis/axisutils';
import { bbox, getBoundingRect, pick } from '../utils/zcutils';
import { POINTER, UNDEFINED } from '../core/zc-constants';

let buttonCount = 0;

export class CustomButton {
    constructor(className, text, button, parent) {
        this.label = text;
        this.button = button;
        this.parent = parent;
        this.className = className;
        this.init();
    }

    init() {
        let button = this,
            padding = getMargin(button.button, 5),
            className = button.className ? '.' + button.className : `.custom-button-${buttonCount++}`,
            group = (button.group = createElement('g', className, [1], button.parent)),
            rect = (button._rect = createElement('rect', `.custom-button-rect`, [1], group)),
            text = (button._text = createElement('text', `.custom-button-text`, [1], group)),
            fixRadius = obj => obj.radius && (obj.rx = obj.ry = obj.radius);

        text.text(button.label.text).attrs({ dy: '.9em' });
        setStyles(button._text, button.label.style);

        let boundingBox = getBoundingRect(text);
        let width = pick(button.button.width, boundingBox.width),
            height = pick(button.button.height, boundingBox.height),
            dimension = {
                width: width + padding.left + padding.right,
                height: height + padding.top + padding.bottom
            },
            textOffset = {
                x: dimension.width / 2 - boundingBox.width / 2,
                y: dimension.height / 2 - boundingBox.height / 2
            };

        // fix the radius
        fixRadius(button.button, dimension);

        //#ZC2153
        setStyles(rect, { ...dimension, ...button.button }, UNDEFINED, { width: 'width', height: 'height' });

        text.attrs(textOffset);
        button.group.styles({ cursor: POINTER });

        //save the bounds
        button.bounds = {
            text: {
                ...textOffset,
                width: boundingBox.width,
                height: boundingBox.height
            },
            rect: dimension
        };
    }

    translate(translate) {
        this.group.attr('transform', `translate(${translate.join(' ')})`);
    }

    show() {
        this.group.style('display', null);
        this.visible = true;
    }

    hide() {
        this.group.style('display', 'none');
        this.visible = false;
    }
}
