// @zohocharts/zc/src/components/conversion-line/conversion-line-parser.js
//
import { MarkerParser } from '../../utils/marker-parser';
import { Registry } from '../../core/zc-registry';
import { getConversionMarkerData } from './conversion-line-adapter';
import { defaultHash } from '../../core/data-processor/flatten-data';

export function conversionLineParser(notesConfig, config = {}, chartObj) {
    // let { markerInfo, labelInfo, display } = conversionLineH;
    let data = notesConfig.data,
        conversionLineH = chartObj.conversionH,
        _data = chartObj.seriesdata[0].data[0];
    // if (!conversionLineH) {
    chartObj.conversionH = conversionLineH = getConversionMarkerData(null, notesConfig.config, chartObj);
    let { eventsInfo } = conversionLineH.markerInfo,
        { highlighter, tooltipContent } = eventsInfo;
    // }

    let { markerInfo, labelInfo, display } = conversionLineH;
    let arrowMark = MarkerParser({
        selection: '.outer',
        data: [data],
        config: {
            x: function(d, i) {
                return markerInfo.x(d, i);
            },
            y: function(d, i) {
                return markerInfo.y(d, i);
            },
            ...markerInfo.options
        },
        event: {
            type: 'zc-marker',
            id: 'cLine',
            superKey: 'segment',
            size: (markerInfo.options.size * 10) / 2,
            data: (_d, chart, xy, parser) => {
                // TODO: work
                let d = [..._d];
                d = defaultHash(d, chart, _d.seriesIndex, _d.itemIndex, 'zc-marker', _d._key);
                d._dependents = {
                    from: _data[parser.conf.prev], // TODO: work
                    to: _data[parser.conf.cur]
                };
                    
                d._data = function() {
                    if (highlighter === undefined) {
                        return [];
                    }
                    let point = highlighter.data().call(this);
                    let actualPoint = chart.annotation.events.find(element => element.data._key === point.__key);

                    point._x = actualPoint.xy[0] - chart.plotarea.left;
                    point._y = actualPoint.xy[1] - chart.plotarea.top;
                    return point;
                };
                return d;
            },
            highlighter,
            tooltipContent,
            ...(eventsInfo.config || {})
        }
    });
    let _conversionLineMark = {
        mark: {
            mark: 'g',
            selector: '.conversion-line',
            data: [1],
            properties: {
                attrs: {
                    // transform: `translate(${chartObj.plotarea.left},${chartObj.plotarea.top})`
                }
            },
            children: [
                {
                    ...arrowMark
                },
                {
                    name: 'label',
                    selector: '.conversion-label',
                    mark: 'text',
                    data: [data],
                    properties: {
                        text: function(d, i) {
                            return labelInfo.formatter(d, notesConfig);
                        },
                        attrs: {
                            x: (d, i) => labelInfo.x(d, i),
                            y: (d, i) => labelInfo.y(d, i),
                            dy: labelInfo.dy
                        },
                        styles: {
                            ...labelInfo.styles
                        }
                    }
                }
            ]
        }
    };
    return _conversionLineMark.mark;
}
Registry.setComponent('conversionLineParser', conversionLineParser);
