// @zohocharts/zc/src/components/conversion-line/conversion-line-component.js
import { renderDom } from '../../mark/dom-render';
import { Registry } from '../../core/zc-registry';
import { getConversionMarkerData } from './conversion-line-adapter';
import { getConversionLineDependents } from './conversion-line-adapter';
import { conversionLineParser } from './conversion-line-parser';
import { defaultHash } from '../../core/data-processor/flatten-data';
import { MARKER_COMPONENT } from '../component-name-constants';

export function conversionLineComponent(data, userConfig, chart) {
    let data_c = data[0].data[0], //chart.seriesdata[0].data[0]
        conversionLineH = getConversionMarkerData(data_c, userConfig, chart),
        marks = conversionLineParser(data_c, conversionLineH, chart),
        eventInfo;

    renderDom(marks.mark, chart.staticPaneSVG.selectAll('g.qualitativeGroupHolder'));

    if (conversionLineH.markerInfo.eventsInfo.config.enabled) {
        let id = `cLine`,
            highlighter = conversionLineH.markerInfo.eventsInfo.highlighter;
        eventInfo = {
            type: MARKER_COMPONENT,
            id,
            config: {
                [id]: data_c.map((_d, i) => {
                    let d = [..._d];
                    d = defaultHash(d, chart, 0, i, MARKER_COMPONENT);
                    d._data = highlighter.data();
                    d._dependents = getConversionLineDependents(d, i, data_c);
                    return d;
                }),
                superKey: 'segment',
                componentIndex: 0
            }
        };
        eventInfo.config[id].disabled = d => false;
        chart.components = [
            {
                name: 'conversionLine',
                events: eventInfo,
                markerH: conversionLineH.markerInfo,
                highlighter
            }
        ];
    }
}

Registry.setComponent('conversionLine', conversionLineComponent);
