// @zohocharts/zc/src/components/conversion-line/conversion-line-adapter.js
import { getOrdinalRange } from '../../axis/scale';
import { applyFont } from '../../utils/zcsvgutils';
import { $ZCG } from '../../core/zcglobal';
import { defined, invokeFunction, mergeJSON } from '../../utils/zcutils';
import { getAnimationConf } from '../../utils/animationutils';
import { mathAbs } from '../../core/zc-constants';
import { Markerhighlighter } from '../marker/marker-highlight';
import { getConversionLineConfig } from './conversion-line-config';
import { MARKER_COMPONENT } from '../component-name-constants';
import { defaultHash } from '../../core/data-processor/flatten-data';

export function getConversionMarkerData(data, userConfig, chart) {
    data = data || chart.seriesdata[0].data[0];
    let defaultConfig = getConversionLineConfig(),
        axisRotated = chart.axes.rotated,
        axisReverse = chart.systemConf.chart.axes.xaxis.reversed,
        m = axisRotated ? -1 : -1,
        xScale = chart.axes.x.scale,
        yScale = chart.axes.y.groups[0].scale,
        bandwidth = Math.min(xScale.bandwidth(),xScale.maxBandWidth),
        ordinalRange = getOrdinalRange(xScale),
        diffBand = xScale.domain().length > 1 ? mathAbs(ordinalRange[1] - ordinalRange[0]) + m * bandwidth : 0,
        bandGap = axisRotated && !axisReverse ? -diffBand : diffBand,
        percentSpacing = 35,
        spacing = ((percentSpacing / 100) * diffBand).toFixed(2),
        isRangeChart = chart.dataObject.noofBarRangeEnabledSeries > 0,
        range = isRangeChart ? 1 : null,
        userArgs = {
            x: xScale,
            y: yScale,
            bandGap,
            bandwidth,
            addExtraBW: axisRotated && !axisReverse ? bandwidth : 0,
            padding: spacing,
            labelPaddingUnit: axisRotated ? 1 : -1,
            isRangeChart,
            diffBand
        },
        customConfig = userConfig ? invokeFunction(userConfig, userArgs) : {},
        config = mergeJSON(defaultConfig, customConfig, true),
        { conversionLine } = config,
        { marker, label } = conversionLine,
        size = defined(marker.size) ? marker.size : diffBand - 2 * spacing,
        angle = marker.angle ? marker.angle : axisRotated ? (axisReverse ? 90 : -90) : axisReverse ? -180 : 0,
        animationOptions = getAnimationConf(
            chart.renderer[0].commonRendererProp().params.commonPlotOptions,
            chart.seriesdata[0].plotoptions,
            chart,
            'bar'
        ),
        x_getter = chart.dataset.getX,
        y_getter = chart.dataset.getY;

    /*
    COMMON INFO
     */
    let display = (d, i) => {
        if (axisReverse && !axisRotated) {
            return i == d.length - 1 ? 'none' : 'show';
        } else {
            return i == '0' ? 'none' : 'show';
        }
    };

    /*
    CONVERSION MARKER INFO
     */
    let x = (d, i) => {
            return axisRotated ? marker.y(chart.dataset.getY(d, 0), i, userArgs) : marker.x(chart.dataset.getX(d), i, userArgs);
        },
        y = (d, i) => {
            return axisRotated ? marker.x(chart.dataset.getX(d), i, userArgs) : marker.y(chart.dataset.getY(d, 0), i, userArgs);
        },
        options = {
            size: size,
            shape: marker.shape,
            fillOpacity: marker.fillOpacity,
            strokeOpacity: marker.strokeOpacity,
            strokeWidth: marker.strokeWidth,
            fillColor: marker.fillColor,
            strokeColor: marker.strokeColor || 'black',
            animation: animationOptions,
            angle
        },
        highlightOptions = {
            segment: axisRotated ? 'line' : 'vertical-line'
        };

    /*
    CONVERSION LABEL INFO
     */
    let defaultFormatConf = {
        format: {
            decimalPlaces: 0,
            suffix: '%'
        }
    };
    let x_label = (d, i) => {
            return axisRotated ? label.y(chart.dataset.getY(d, 0), i, userArgs) : label.x(chart.dataset.getX(d), i, userArgs);
        },
        y_label = (d, i) => {
            return axisRotated ? label.x(chart.dataset.getX(d), i, userArgs) : label.y(chart.dataset.getY(d, 0), i, userArgs);
        },
        styles = applyFont(null, [label, chart.systemConf.chart, chart.systemConf.canvas, $ZCG]),
        formatter = (d, conf) => {
            let current = chart.dataset.getY(d, 0, range),
                previous = conf.cur == 0 ? current : chart.dataset.getY(data[conf.cur - 1], 0, range),
                base = conf.cur == 0 ? current : chart.dataset.getY(data[0], 0, range);

            return label.text({ current, previous, base }, conf.cur, userArgs);
        };
    styles['text-anchor'] = axisRotated ? 'start' : 'middle';
    let labelInfo = { x: x_label, y: y_label, formatter, styles };

    /*
    MARKER EVENTS INFO
     */
    let eventsInfo = {};
    let highlighter = new Markerhighlighter(chart, { x, y, options, highlightOptions });
    highlighter.options({ options });
    if (marker.events.enabled) {
        eventsInfo = {
            config: marker.events,
            highlighter,
            tooltipContent: (d, i) => {
                let currentX = chart.dataset.getX(d, 0, range),
                    previousX = i == 0 ? currentX : chart.dataset.getX(data[i - 1], 0, range),
                    nextX = i == data.length - 1 ? currentX : chart.dataset.getX(data[i + 1], 0, range);

                if (axisReverse) {
                    return `${currentX}->${nextX}: ${formatter(d, i, data)}`;
                }

                return `${previousX}->${currentX}: ${formatter(d, i, data)}`;
            }
        };
    }

    let markerInfo = { x, y, options, eventsInfo };

    //
    // let id = `cLine`;
    // let eventInfo = {
    //     type: MARKER_COMPONENT,
    //     id,
    //     config: {
    //         [id]: data.map((_d, i) => {
    //             let d = [..._d];
    //             d = defaultHash(d, chart, 0, i, MARKER_COMPONENT);
    //             d._data = highlighter.data();
    //             d._dependents = getConversionLineDependents(d, i, data);
    //             return d;
    //         }),
    //         superKey: 'segment',
    //         componentIndex: 0
    //     }
    // };
    // eventInfo.config[id].disabled = d => false;
    // chart.components = [
    //     {
    //         name: 'conversionLine',
    //         events: eventInfo,
    //         markerH: markerInfo,
    //         highlighter
    //     }
    // ];

    return { markerInfo, labelInfo, display };
}

export function getConversionLineDependents(d, i, data) {
    if (i > 0) {
        return {
            from: data[i - 1],
            to: data[i]
        };
    }
    return null;
}
