// @zohocharts/zc/src/components/conversion-line/conversion-line-config.js
import { getFormattedValue } from '../../formatter/formatter';
import { NUMERIC_DATATYPE } from '../../core/datatype-constants';

export function getConversionLineConfig() {
    let defaultFormatConf = {
            format: {
                decimalPlaces: 0,
                suffix: '%'
            }
        },
        config = {
            conversionLine: {
                marker: {
                    shape: 'arrow-line',
                    fillOpacity: 1,
                    strokeOpacity: 1,
                    strokeWidth: 1,
                    strokeColor: 'grey',
                    fillColor: 'none',
                    x: (d, i, args) => {
                        return -args.bandGap / 2 + args.addExtraBW;
                    },
                    y: (d, i, args) => {
                        return 0;
                    },
                    events: {
                        enabled: true
                        
                    }
                },
                label: {
                    x: (d, i, args) => {
                        return -args.bandGap / 2 + args.addExtraBW;
                    },
                    y: (d, i, args) => {
                        return args.labelPaddingUnit * 20;
                    },
                    text: (values, i, args) => {
                        let percentCoversion = (values.current / values.base) * 100;
                        return getFormattedValue(defaultFormatConf, percentCoversion, NUMERIC_DATATYPE);
                    }
                }
            }
        };
    return config;
}
