// @zohocharts/zc/src/scroll/defaultScroll.js
//$Id$
import { DIV_ELEMENT } from '../core/element-constants';
import { DIAL_CHARTNAME } from '../core/chartname-constants';
import { Registry } from '../core/zc-registry';
import { chartcategory, renderer_helpers } from '../core/main-zc-init';
import { getSvgTransform } from '../utils/svg-transforms';
import { appendEle, createElement } from '../utils/zcsvgutils';
import { getAxisObject, getAxisAreaForLowerOrder } from '../axis/axisutils';
import { getPlotSize } from '../plot/plotSize';
import { PX, TRANSPARENT, NONE, mathMax, mathMin } from '../core/zc-constants';
import { d3_event, d3_select } from '../core/d3-constants';
import { defined, arrayIncludes, invokeFunction } from '../utils/zcutils';
import { svg_clips } from '../effects/svg-clips';
import { $Browser } from '../core/browser-detect';
import { isLeafAxis, isFacet } from '../axis/hierarchy-axis-utils';
import { seriesIs } from '../events/events-utils';

export var plotScroll_div = function() {};
plotScroll_div.prototype.init = function(chartObj){
    if (!isSmoothScrollEnabled(chartObj)) {
        return;
    }
    var axisrotated = chartObj.axes.rotated,
        plotarea = chartObj.plotarea,
        XSIZE = getPlotSize(chartObj, 'width') - plotarea.width,
        YSIZE = getPlotSize(chartObj, 'height') - plotarea.height;

    this.chart_Width = invokeFunction(chartObj.systemConf.chart.width, chartObj);
    this.chart_Height = invokeFunction(chartObj.systemConf.chart.height, chartObj);
    this.scroll_Width = plotarea.width + (this.mode == 'xy' || this.mode == 'x' || !axisrotated ? XSIZE : 0);
    this.scroll_Height = plotarea.height + (this.mode == 'xy' || this.mode == 'y' || axisrotated ? YSIZE : 0);

    chartObj.defaultScroll.dimension = {
        width: this.chart_Width === null || isNaN(this.chart_Width) ? this.chart_Width : this.scroll_Width,
        height: this.chart_Height === null || isNaN(this.chart_Height) ? this.chart_Height : this.scroll_Height
    };
}
plotScroll_div.prototype.defaultScroll = function(chartObj, state) {
    if (!isSmoothScrollEnabled(chartObj)) {
        return;
    }
    // this.init(chartObj);
    var isWidth = defined(this.chart_Width),
        isHeight = defined(this.chart_Height);
    var mode = (this.mode = isWidth && isHeight ? 'xy' : isWidth ? 'x' : 'y');
    var defaultScroll = this,
        axisrotated = chartObj.axes.rotated,
        plotarea = chartObj.plotarea,
        viewParserNew = chartObj.viewFlow && chartObj.viewFlow.spec.renderOldConfig == false,
        isSupported = viewParserNew ? false : arrayIncludes(chartcategory.nonAxisSingleSourceEvent, chartObj.chartTypes[0].value);

    var XSIZE = getPlotSize(chartObj, 'width') - plotarea.width,
        YSIZE = getPlotSize(chartObj, 'height') - plotarea.height;

    // Initial clip-path
    var scroll = chartObj.container.selectAll('.d3container').selectAll('.zohocharts_scroll'),
        scrollbar_size = +scroll.attr('scrollbarSize');
    
    let scrollWidth = this.scroll_Width,
    scrollHeight = this.scroll_Height;

    chartObj.defaultScrollWidth = scrollWidth;

    updateAxisClipPath(chartObj, mode, 'init');

    let scrollStyles = {
        left: plotarea.left + PX,
        top: plotarea.top + PX,
        height: plotarea.height + (this.mode == 'xy' || this.mode == 'x' ? scrollbar_size : 0) + PX,
        width: plotarea.width + (this.mode == 'xy' || this.mode == 'y' ? scrollbar_size : 0) + PX,
        'overflow-y': this.mode == 'xy' || this.mode == 'y' || axisrotated ? 'auto' : 'hidden',
        'overflow-x': this.mode == 'xy' || this.mode == 'x' || !axisrotated ? 'auto' : 'hidden'
    };

    // ZC4087 - arrow marks  is not in proper position after scrolling and redraw
    chartObj.container.selectAll('.qualitativeGroupHolder').selectAll('g#plot_notes_main').attrs({_x: null, _y: null});

    // Scroll div
    scroll.styles(scrollStyles).on('scroll', function(e) {
        const d3event = d3_event(e);
        var src = d3event.srcElement || d3event.target,
            scrolled;
        // var scrolled = mode == 'xy' ? src.scrollLeft || src.scrollTop : !axisrotated ? src.scrollLeft : src.scrollTop;

        var scrollLeft = src.scrollLeft,
            scrollTop = src.scrollTop;

        var isScrollX = chartObj.scrolledX !== scrollLeft,
            isScrollY = chartObj.scrolledY !== scrollTop;

        chartObj.scrolledY = scrollTop;
        chartObj.scrolledX = scrollLeft;

        var scrollingMode = isScrollX ? 'x' : 'y';

        scrolled = isScrollX ? scrollLeft : scrollTop;

        //#ZC974
        d3event.clientX = d3event.clientX || 0;
        d3event.clientY = d3event.clientY || 0;

        // hide the tooltip when scrolled
        if (chartObj.dataObject.isAxisCategory) {
            chartObj.eventHandler.tooltipElement.hide();
            chartObj.eventHandler.handlers.mouseout.call(chartObj.eventHandler);
        }

        //Reset cache for data
        chartObj.cache.data = {};
        chartObj.cache.search = {};

        scrollGroups(scrolled, mode, scrollingMode);
    });

    if (chartObj.canvasScrollGroup) {
        chartObj.canvasScrollGroup.styles(scrollStyles);
    }

    // fixes #ZC1767 (ticks hiding chrome)
    if ($Browser.CHROME) {
        var canvasArea = chartObj.canvasarea;
        defaultScroll.fixScrollView(
            false,
            {
                width: canvasArea.width + (this.mode == 'xy' || !axisrotated ? XSIZE : 0),
                height: canvasArea.height + (this.mode == 'xy' || axisrotated ? YSIZE : 0)
            },
            chartObj
        );
    }

    scroll.selectAll('.child').styles({
        height: this.scroll_Height + PX,
        width: this.scroll_Width + PX,
        'pointer-events': NONE,
        margin: 0,
        padding: 0,
        'background-color': TRANSPARENT
    });

    // Reset scroll position
    //ZC1768 (why reset?)
    // if (state !== "addorremoveseries") {
    // var scroll_node = scroll.node();
    // scroll_node.scrollLeft = 0;
    // scroll_node.scrollTop = 0;
    // }

    // Event listners
    if (chartObj.dataObject.isAxisCategory || chartObj.dataObject.isFlowChartCategory || isSupported) {
        chartObj.eventHandler.addListeners(scroll);
    }

    var scrollingMode = chartObj.scrolledX ? 'x' : 'y';
    var scrolled_1 = scrollingMode == 'x' ? chartObj.scrolledX : chartObj.scrolledY;

    if (scrolled_1 && state === 'addorremoveseries') {
        scrollGroups(scrolled_1, this.mode, scrollingMode);
        (chartObj.scrolledX && chartObj.scrolledY) ? scrollGroups(chartObj.scrolledY, this.mode, 'y') : null;
    }

    function scrollGroups(scrolled, mode, scrollingMode) {
        var seriesholder = chartObj.container.selectAll('.seriesholder'),
            seriesHolderClipId = chartObj.container.select('.seriesClippathholder').attr('clip-path'),
            seriesTrans = getSvgTransform(seriesholder).translate,
            transX = scrollingMode == 'x' ? plotarea.left - scrolled : seriesTrans[0],
            transY = scrollingMode == 'x' ? seriesTrans[1] : plotarea.top - scrolled,
            mapping_orient = {
                x: axisrotated ? 'height' : 'width',
                y: axisrotated ? 'width' : 'height'
            },
            mapping_axistype = {
                x: axisrotated ? 'y' : 'x',
                y: axisrotated ? 'x' : 'y'
            };
        
        const { scrolledX = 0, scrolledY = 0 } = chartObj;

        seriesholder.attr('transform', `translate(${transX},${transY})`);
        chartObj.container.selectAll('.zc-fade-overlay').attr('transform', `translate(${-chartObj.scrolledX},${-chartObj.scrolledY})`);
        chartObj.container.selectAll('.labelholder').attr('transform', `translate(${transX},${transY})`);
        chartObj.container.selectAll('.eventgroup').attr('transform', `translate(${-chartObj.scrolledX},${-chartObj.scrolledY})`);
        chartObj.container.selectAll('rect.eventhandler').attr('transform', `translate(${chartObj.scrolledX},${chartObj.scrolledY})`);
        chartObj.container.selectAll('.qualitativeGroup').attr('transform', `translate(${-chartObj.scrolledX},${-chartObj.scrolledY})`);
        // chartObj.notes.selectAll('#plot > div').style('transform', `translate(${-chartObj.scrolledX + PX},${-chartObj.scrolledY + PX})`); //ZC4225

        // #ZC1770
        let translate = `translate(${-chartObj.scrolledX + PX},${-chartObj.scrolledY + PX})`;
        chartObj.container.selectAll('canvas').style('transform', translate);
        // #ZC1840 IE not scrolling
        chartObj.container.selectAll('canvas').style(`${$Browser.prefix.css}transform`, translate);
        chartObj.container.selectAll('.highlight-canvas').style('transform', null); // ZC4238

        chartObj.drawPlotareaClippath();

        if(chartObj.notesParent && chartObj.notesParent.length){
            chartObj.notesParent.map(element => {
                if(element.node().ownerSVGElement){
                    let __element = element.selectAll("g#plot_notes_main");
                    let _x = __element.attr('_x'),
                    _y = __element.attr('_y'), _transX, _transY;
                
                    if(defined(_x)){
                        
                        _transX = scrollingMode == 'x' ? parseFloat(_x) - scrolled : parseFloat(_x) - scrolledX,
                        _transY = scrollingMode == 'x' ? parseFloat(_y) - scrolledY : parseFloat(_y) - scrolled;
    
                    }else{
                        let _seriesTrans = getSvgTransform(__element).translate;
    
                        __element.attrs({
                            _x: _seriesTrans[0],
                            _y: _seriesTrans[1]
                        })
    
                        _transX = scrollingMode == 'x' ? _seriesTrans[0] - scrolled : _seriesTrans[0] - scrolledX,
                        _transY = scrollingMode == 'x' ? _seriesTrans[1] - scrolledY : _seriesTrans[1] - scrolled;
                    }
    
                    __element.attr('transform', `translate(${_transX},${_transY})`);

                    element.attr('clip-path', `url(${chartObj.defsLocation}#${chartObj.plotClippathID})`)
                }
                
            })
        }
        
        var attrObj =
            scrollingMode == 'y'
                ? {
                      y: -scrolled,
                      height: plotarea.height + scrolled
                  }
                : {
                      x: -scrolled,
                      width: plotarea.width + scrolled
                  };

        if (
            chartObj.dataObject.isAxisCategory ||
            chartObj.dataObject.isPolarAxisCategory ||
            chartObj.dataObject.isFlowChartCategory ||
            isSupported
        ) {
            chartObj.eventHandler.trackerElement.attrs(attrObj).attr('clip-path', seriesHolderClipId);
        }

        updateAxisClipPath(chartObj, scrollingMode);

        if (chartObj.dataObject.isFlowChartCategory) {
            renderer_helpers.sankey.scrollClipPath(chartObj, chartObj.wrap.selectAll('g.seriesholder'));
        }

        var properties = {
                name: 'rect_highlight',
                id: chartObj.id
            },
            args = {
                chartObj: chartObj
            };
        if (scrollingMode == 'y') {
            properties.x = -plotarea.left + chartObj.scrolledX;
            properties.y = scrolled;
            properties.width = plotarea.width + plotarea.left;
            properties.height = plotarea.height;
        } else {
            properties.x = scrolled;
            properties.y = plotarea.top + chartObj.scrolledY;
            properties.width = plotarea.width;
            properties.height = plotarea.height;
        }

        // #ZC1769
        var clipId = `default_scroll_highlight_clip_${chartObj.id}`;
        svg_clips.rect(properties, clipId, null, null, null, chartObj);

        //BG Image position update
        if (chartObj.plot.hasBgAspectRatio) {
            let bgEle = chartObj.plotareaPropertyElement;
            bgEle.styles({
                'background-position-x': `${-chartObj.scrolledX}px`,
                'background-position-y': `${-chartObj.scrolledY}px`
            });
        }

        // var clipID = effects.get(effects.svg_clips, "rect", properties, args);
        // chartObj.container.selectAll('#zcareahighlight').attr('clip-path', `url(#${clipId})`);
    }
};

plotScroll_div.prototype.initialize = function(chartObj) {
    if (!isSmoothScrollEnabled(chartObj)) {
        return;
    }
    var isWidth = defined(chartObj.systemConf.chart.width),
        isHeight = defined(chartObj.systemConf.chart.height);
    var mode = (this.mode = isWidth && isHeight ? 'xy' : isWidth ? 'x' : 'y');

    var axisrotated = chartObj.axes.rotated,
        mapping_orient = {
            x: axisrotated ? 'height' : 'width',
            y: axisrotated ? 'width' : 'height'
        },
        mapping_axistype = {
            x: axisrotated ? 'y' : 'x',
            y: axisrotated ? 'x' : 'y'
        };

    var plotarea = chartObj.plotarea;

    chartObj.scrolledX = chartObj.scrolledX || 0;
    chartObj.scrolledY = chartObj.scrolledY || 0;

    var scrollStyles = {
        overflow: 'scroll',
        position: 'absolute',
        margin: 0,
        padding: 0,
        'background-color': 'rgba(0,0,0,0)',
        'box-sizing': 'content-box',
        '-webkit-tap-highlight-color': TRANSPARENT,
        '-webkit-overflow-scrolling': 'touch'
    };

    var scroll = appendEle(chartObj.container.selectAll('.d3container'), DIV_ELEMENT, [1], '.', 'class', 'zohocharts_scroll');
    scroll.styles(scrollStyles);

    appendEle(scroll, DIV_ELEMENT, [1], '.', 'class', 'child');

    var axisrotated = chartObj.axes.rotated,
        conf = chartObj.defaultScroll;

    //scrollbar size
    var scroll_node = scroll.node();
    var scrollbar_size;
    if (mode == 'xy') {
        scrollbar_size = scroll_node.offsetWidth - scroll_node.clientWidth + (scroll_node.offsetHeight - scroll_node.clientHeight);
    }
    var scrollbar_size =
        axisrotated || mode == 'y'
            ? scroll_node.offsetWidth - scroll_node.clientWidth
            : scroll_node.offsetHeight - scroll_node.clientHeight;
    scroll.attr('scrollbarSize', scrollbar_size);

    var axistype_horizontal = mapping_axistype.x,
        axistype_vertical = mapping_axistype.y,
        axisObj_horizontal = axistype_horizontal == 'x' ? chartObj.systemConf.chart.axes.xaxis : chartObj.systemConf.chart.axes.yaxis[0],
        axisObj_vertical = axistype_vertical == 'x' ? chartObj.systemConf.chart.axes.xaxis : chartObj.systemConf.chart.axes.yaxis[0];
    if (!conf) {
        conf = chartObj.defaultScroll = {};

        conf.tickmarkSize = +axisObj_horizontal.tickmark.size;
        conf.marginRight = +chartObj.systemConf.chart.marginRight;
        conf.tickmarkSizeVertical = +axisObj_vertical.tickmark.size;
    }
    if ((mode == 'x' || mode == 'xy') && axisObj_horizontal.orient !== 'top') {
        axisObj_horizontal.tickmark.size = +conf.tickmarkSize + scrollbar_size;
    }
    if ((mode == 'y' || mode == 'xy') && axisObj_vertical.orient == 'right') {
        axisObj_vertical.tickmark.size = +conf.tickmarkSizeVertical + scrollbar_size;
    }

    // #ZC1770
    // add canvas group for canvas charts
    if (chartObj.systemConf.chart.plot.renderer.mode.toLowerCase() === 'canvas') {
        chartObj.canvasScrollGroup = createElement('div', '.canvas_scroll_group', [1], chartObj.chartParentElement, '.static').styles(
            scrollStyles
        );
    }

    if (mode == 'xy' || mode == 'y') {
        chartObj.plotarea.bottom += scrollbar_size;
        chartObj.plotarea.width -= scrollbar_size;
    }
};

function getClipPathAttrs(chartObj, axisType, dimension, axisOrder) {
    var hierachyObj = chartObj.dataObject.hierarchical,
        plotarea = chartObj.plotarea,
        scrolled = dimension == 'width' ? chartObj.scrolledX : chartObj.scrolledY,
        attrObj,
        axesCount = chartObj.axes[axisType].count;
    var axisLen = 0;
    if (hierachyObj) {
        axisLen = getAxisAreaForLowerOrder(axisType, axisOrder, chartObj);
    } else {
        axisLen = axisType == 'x' ? chartObj.axes.x.ticks.preCalcAxisLen : chartObj.axes.y.groups[0].ticks.preCalcAxisLen;
    }

    let axisIndex = axisOrder;
    if (isFacet(chartObj) && !chartObj.isAxisCategory) {
        // Pivot non-axis chart
        axisIndex = hierachyObj.pivotLayout.getRootNodeList(axisType == 'y' ? 'row' : 'column').getLeavesCount();
    }

    var axisObj_cur = getAxisObject(chartObj, axisType, axisIndex),
        axisLen_cur = axisObj_cur.component.tickLabelWidth(),
        orient = axisObj_cur.component.orient();

    //margin percent applied - reference from axis-clippath.js
    var percentLimit = 0,
        margin = {
            left: percentLimit * chartObj.plotarea.left,
            right: percentLimit * chartObj.plotarea.right,
            top: percentLimit * chartObj.plotarea.top,
            bottom: percentLimit * chartObj.plotarea.bottom
        }
    if (dimension == 'height') {
        if (orient == 'left') {
            attrObj = {
                x: -plotarea.left,
                y: scrolled - margin.top,
                width: plotarea.width + plotarea.left + axisLen,
                height: plotarea.height + margin.top + margin.bottom
            };
        } else {
            attrObj = {
                x: -axisLen - plotarea.width,
                y: scrolled - margin.top,
                width: plotarea.width + axisLen_cur + axisLen,
                height: plotarea.height + margin.top + margin.bottom 
            };
        }

        var percentLimit = 0.5,
            margin = {
                left: percentLimit * chartObj.plotarea.left,
                right: percentLimit * chartObj.plotarea.right,
                top: percentLimit * chartObj.plotarea.top,
                bottom: percentLimit * chartObj.plotarea.bottom
            }
        
        if((chartObj.userdata.chart.height == 'auto' && chartObj.systemConf.chart['height'] == 'auto') || chartObj.userdata.chart.height < plotarea.height){
            attrObj.y = scrolled - margin.top;
            attrObj.height = plotarea.height + margin.top + margin.bottom;
        }
        else{
            if(scrolled == 0){
                attrObj.y = scrolled - margin.top;
                attrObj.height = plotarea.height + margin.top;
            }
            else if(defined(chartObj.defaultScroll.dimension)){
                var scrollPercentage = (100 * (scrolled + plotarea.height)/chartObj.defaultScroll.dimension.height);
                if(scrollPercentage > 99.8){
                    attrObj.y = scrolled;
                    attrObj.height = plotarea.height + margin.bottom;
                }
            }
        }
    } 
    else {
        if (orient == 'bottom') {
            attrObj = {
                x: scrolled,
                y: -axisLen - plotarea.height,
                width: plotarea.width,
                height: axisLen + plotarea.height + axisLen_cur
            };
        } else {
            attrObj = {
                x: scrolled,
                y: -plotarea.top,
                width: plotarea.width,
                height: plotarea.height + plotarea.top + axisLen
            };
        }
        //Extending margin ZC3789
        var percentLimit = 0.5,
            margin = {
                left: percentLimit * chartObj.plotarea.left,
                right: percentLimit * chartObj.plotarea.right,
                top: percentLimit * chartObj.plotarea.top,
                bottom: percentLimit * chartObj.plotarea.bottom
        }
        if((scrolled + plotarea.width == chartObj.defaultScrollWidth) || chartObj.userdata.chart.width < plotarea.width){
            attrObj.x = scrolled - margin.left;
            attrObj.width = plotarea.width + margin.left + (2 * margin.right);
        }
        else{ 
            if(scrolled == 0){
                attrObj.x = scrolled - margin.left;
                attrObj.width = plotarea.width + margin.left;
            }
            else if(defined(chartObj.defaultScroll.dimension)){
                var scrollPercentage = (100 * (scrolled + plotarea.width)/chartObj.defaultScroll.dimension.width);
                if(scrollPercentage > 99.8){  
                    attrObj.x = scrolled;
                    attrObj.width = plotarea.width + (2 * margin.right);
                }
            }
        }
    }
    return attrObj;
}

function getPolarClipPathAttrs(chartObj, axisType, node) {
    let scrollPosByEvent = getScrollPositionByEvent(chartObj),
        trans = getSvgTransform(node.parentNode).translate,
        plotarea = chartObj.plotarea,
        plotarea_approx = chartObj.plotarea_approx,
        bound = chartObj.rendererConf[0].bound(),
        polarMode =
            (chartObj.dataObject.isPolarAxisCategory && axisType === 'x') || (!chartObj.dataObject.isPolarAxisCategory && axisType === 'y');

    let x = polarMode
            ? -trans[0] - plotarea_approx.x + plotarea.left + scrollPosByEvent[0]
            : -trans[0] + plotarea.left + scrollPosByEvent[0],
        y = polarMode ? -trans[1] - plotarea_approx.y + plotarea.top + scrollPosByEvent[1] : -trans[1] + plotarea.top + scrollPosByEvent[1],
        width = mathMin(plotarea.width, mathMax(-x + bound.width, 0)),
        height = mathMin(plotarea.height, mathMax(-y + bound.height, 0));

    return { x, y, width, height };
}

function getAxisGroups(chartObj, axisType, dimension, leafOnly) {
    var axis = chartObj.container.selectAll(`g.${axisType}-axis`).selectAll('g.axisgroup,g.minmaxlabelsGroup'),
        [transX, transY] = getScrollPosition(chartObj, dimension),
        scrollPosByEvent = getScrollPositionByEvent(chartObj),
        isPolarLeafAxis = _class => _class && _class.search('-') > -1;

    axis.each(function(d, i) {
        var comp = d3_select(this),
            _class,
            clipAttrs;

        if (hasPolarAxis(chartObj)) {
            _class = getAxisID(axisType, this);

            if (comp.attr('class') === 'minmaxlabelsGroup' || isPolarLeafAxis(_class)) {
                // Leaf axis
                clipAttrs = getPolarClipPathAttrs(chartObj, axisType, this);
            } else {
                if (leafOnly) {
                    return;
                }

                clipAttrs = getClipPathAttrs(chartObj, axisType, dimension, +_class);
            }
        } else {
            clipAttrs = getClipPathAttrs(chartObj, axisType, dimension, i);
        }

        let pos = isPolarLeafAxis(_class)
            ? (chartObj.dataObject.isPolarAxisCategory && axisType === 'x') ||
              (!chartObj.dataObject.isPolarAxisCategory && axisType === 'y')
                ? [-scrollPosByEvent[0] + chartObj.plotarea_approx.x, -scrollPosByEvent[1] + chartObj.plotarea_approx.y]
                : [-scrollPosByEvent[0], -scrollPosByEvent[1]]
            : [transX, transY];
        var clipid = `default_scroll_axes_${axisType}_${i}_clip${chartObj.id}`;
        svg_clips.rect(clipAttrs, clipid, null, null, null, chartObj);

        comp.attrs({
            transform: `translate(${pos[0]},${pos[1]})`,
            'clip-path': `url(#${clipid})`
        });
        chartObj.staticPaneSVG.selectAll('g.threshold').attrs({
            transform: `translate(${pos[0]},${pos[1]})`
        });
        if (chartObj.dataObject.hierarchical && chartObj.dataObject.hierarchical.categoryTree && !isPolarLeafAxis(_class)) {
            var isLeafAxisType = isLeafAxis(axisType, i, chartObj);
            if (isLeafAxisType) {
                var axislabelG = d3_select(this.parentElement)
                    .selectAll('g.axislabelgroup')
                    .attrs({
                        transform: `translate(${transX},${transY})`,
                        'clip-path': `url(#${clipid})`
                    });
            }
        }

        // comp.selectAll('g.axisgroup').attr('clip-path', '');
    });
}

function getAxisID(axisType, node) {
    return d3_select(node.parentNode)
        .attr('class')
        .replace(`${axisType}-axis-`, '');
}

function getScrollPosition(chartObj, dimension) {
    let scrolled = dimension == 'width' ? chartObj.scrolledX : chartObj.scrolledY;
    if (dimension == 'height') {
        return [0, -scrolled];
    } else {
        return [-scrolled, 0];
    }
}

function getScrollPositionByEvent(chartObj) {
    let event = d3_event();
    if (event) {
        var src = event.srcElement || event.target;
        return [src.scrollLeft, src.scrollTop];
    }

    return [chartObj.scrolledX || 0, chartObj.scrolledY || 0];
}

function updateAxisClipPath(chartObj, scrollingMode, state) {
    var axisrotated = chartObj.axes.rotated,
        plotarea = chartObj.plotarea,
        scrolled = scrollingMode == 'x' ? chartObj.scrolledX : chartObj.scrolledY,
        attrObj,
        hierachyObj = chartObj.dataObject.hierarchical,
        mapping_orient = {
            x: axisrotated ? 'height' : 'width',
            y: axisrotated ? 'width' : 'height'
        },
        mapping_axistype = {
            x: axisrotated ? 'y' : 'x',
            y: axisrotated ? 'x' : 'y'
        };

    if (scrollingMode == 'xy') {
        //x
        getAxisGroups(chartObj, 'x', mapping_orient.x);

        //y
        getAxisGroups(chartObj, 'y', mapping_orient.y);
    } else if (scrollingMode == 'x') {
        //x
        getAxisGroups(chartObj, mapping_axistype[scrollingMode], mapping_orient[mapping_axistype[scrollingMode]]);

        if (hasPolarAxis(chartObj)) {
            getAxisGroups(chartObj, mapping_axistype.y, mapping_orient[mapping_axistype[scrollingMode]], true);
        }
    } else {
        //y
        getAxisGroups(chartObj, mapping_axistype[scrollingMode], mapping_orient[mapping_axistype[scrollingMode]]);

        if (hasPolarAxis(chartObj)) {
            getAxisGroups(chartObj, mapping_axistype.x, mapping_orient[mapping_axistype[scrollingMode]], true);
        }
    }

    if (state == 'init' && scrollingMode == 'xy' && hierachyObj) {
        var rowList = hierachyObj.pivotLayout.getRootNodeList('row'),
            rowCount = rowList.getCount(),
            yaxesCount = chartObj.axes.y.count;
        
        if(!rowCount){
            return
        }
        var parentRowAxisObj = chartObj.axes.y.branches[rowCount - 1],
            gridInfo = parentRowAxisObj.component.majorTicks(),
            orient = parentRowAxisObj.component.orient();

        var scrollGridGroup = createElement(
                'g',
                '.scroll_roller_grid_group',
                [1],
                chartObj.wrap.selectAll(`g.y-axis-${yaxesCount - 1}`),
                '.y-axis'
            ),
            scrollGridlineHeaderGroup = createElement('g', '.scroll_roller_header_grid_group', [1], scrollGridGroup, '.y-axis'),
            scrollGridlineFooterGroup = createElement('g', '.scroll_roller_footer_grid_group', [1], scrollGridGroup, '.y-axis'),
            gridlineNode = chartObj.wrap
                .selectAll(`g.y-axis-${yaxesCount - 1}`)
                .selectAll('line.majorTicks')
                .nodes();
        if (gridlineNode.length) {
            let gridTemplate = gridlineNode[0].cloneNode(0),
                gridTemplate1 = gridlineNode[0].cloneNode(0);

            scrollGridlineHeaderGroup.node().appendChild(gridTemplate);
            scrollGridlineFooterGroup.node().appendChild(gridTemplate1);

            if (orient == 'left') {
                scrollGridlineHeaderGroup.attr('transform', 'translate(0,0.5)');
                scrollGridlineFooterGroup.attr('transform', `translate(0,${chartObj.plotarea.height + 0.5})`);
            } else if (orient == 'bottom') {
                scrollGridlineHeaderGroup.attr('transform', 'translate(0.5,0)');
                scrollGridlineFooterGroup.attr('transform', `translate(${chartObj.plotarea.width + 0.5},0)`);
            }
        }
    }
}

/**
 *
 * @param {boolean} reset
 * @param {{width: number, height: number}} dimension
 * @param {string} chart
 */
plotScroll_div.prototype.fixScrollView = function(reset = false, dimension = {}, chart) {
    chart.svg.styles({
        width: dimension.width + PX,
        height: dimension.height + PX
    });
    chart.d3container.style('overflow', reset ? null : 'hidden');
};

function hasPolarAxis(chartObj) {
    return chartObj.dataObject.isPolarAxisCategory || seriesIs(chartObj.chartTypes[0].value, DIAL_CHARTNAME);
}

export function isSmoothScrollEnabled(chartObj) {
    return (
        ((chartObj._isAxisCategory() || chartObj.dataObject.isFlowChartCategory) && chartObj.systemConf.chart.width) ||
        chartObj.systemConf.chart.height
    );
}
Registry.setComponent('scroll.isSmoothScrollEnabled', isSmoothScrollEnabled);

Registry.setComponent('smoothScroll', plotScroll_div);
