// @zohocharts/zc/src/loader/loader.js
// ;//$Id$
//loader.js
import { pInt } from '../utils/zcutils';
import { _imageFilePath } from '../core/zcconf';
/*
 types = zBlocks, zBars, zCircle1, zCircle2, zCrossBars
 scale - transform scale
 duration - time duration for one iteration
 */
//detect Browser Func

var loadDetectBrowser = function() {
    var userAgent = navigator.userAgent;
    var browser = {};
    if (userAgent.indexOf('OPR') > 0) {
        browser.name = 'Opera';
        browser.jsPrefix = 'webkit';
        browser.cssPrefix = '-webkit-';
        var index = userAgent.indexOf('OPR/');
        browser.version = pInt(userAgent.substr(index + 4));
    } else if (userAgent.indexOf('Chrome') > 0) {
        browser.name = 'Chrome';
        browser.jsPrefix = 'webkit';
        browser.cssPrefix = '-webkit-';
        var index = userAgent.indexOf('Chrome/');
        browser.version = pInt(userAgent.substr(index + 7));
    } else if (userAgent.indexOf('Safari') > 0) {
        browser.name = 'Safari';
        browser.jsPrefix = 'webkit';
        browser.cssPrefix = '-webkit-';
        var index = userAgent.indexOf('Version/');
        browser.version = pInt(userAgent.substr(index + 8));
    } else if (userAgent.indexOf('Firefox') > 0) {
        browser.name = 'Firefox';
        browser.jsPrefix = 'Moz';
        browser.cssPrefix = '-moz-';
        var index = userAgent.indexOf('Firefox/');
        browser.version = pInt(userAgent.substr(index + 8));
    } else if (userAgent.indexOf('MSIE') > 0) {
        browser.name = 'Internet Explorer';
        browser.jsPrefix = 'ms';
        browser.cssPrefix = '-ms-';
        var index = userAgent.indexOf('MSIE');
        browser.version = pInt(userAgent.substr(index + 5));
    } else {
        browser = undefined;
    }
    return browser;
};

var loadSetStyle = function(elementObj, paramsObj) {
    var docStyles = getComputedStyle(elementObj);
    var browser = loadDetectBrowser();

    for (var property in paramsObj) {
        if (paramsObj.hasOwnProperty(property)) {
            if (property in docStyles) {
                elementObj.style[property] = paramsObj[property];
            } else if (browser != undefined) {
                var propertyFix = browser.jsPrefix + property.charAt(0).toUpperCase() + property.slice(1);
                elementObj.style[propertyFix] = paramsObj[property];
            }
        }
    }
};

export var ZCloader = function(elemObj, customObj) {
    //Base Object with default configuration
    var baseObj = {
            type: 'zBars',
            scale: 1,
            duration: 1.2
        },
        style = document.getElementById('ZCloader');

    if (!ZCloader.style) {
        style = document.createElement('style');
        style.setAttribute('id', 'ZCloader');
        document.head.appendChild(style);
        ZCloader.style = true;
    }

    for (var config in customObj) {
        if (customObj.hasOwnProperty(config)) {
            baseObj[config] = customObj[config];
        }
    }

    var clr = baseObj.colors || ['rgb(204, 32, 50)', 'rgb(24, 153, 74)', 'rgb(1, 115, 178)', 'rgb(253, 208, 25)'];
    var clr1 = clr[0 % clr.length],
        clr2 = clr[1 % clr.length],
        clr3 = clr[2 % clr.length],
        clr4 = clr[3 % clr.length];

    var browser = loadDetectBrowser();
    var animString = '';
    if (browser === undefined) {
        browser = {};
        browser.cssPrefix = '';
    } else if (browser.name === 'Internet Explorer' && browser.version < 10) {
        baseObj.type = 'wheel';
    }

    if (!elemObj.loader) {
        var animBlock = document.createElement('div');

        if (baseObj.type == 'zBlocks') {
            if (!ZCloader.zBlocks) {
                var animString = `\n@${browser.cssPrefix}keyframes ${baseObj.type} {`;
                animString += `0% { left:0px; top:0px; ${browser.cssPrefix}transform:rotate(0deg) scale(1)}`;
                animString += `25% { left:40px; top:0px; ${browser.cssPrefix}transform:rotate(90deg) scale(0.5)}`;
                animString += `50% { left:40px; top:40px; ${browser.cssPrefix}transform:rotate(180deg) scale(1.0)}`;
                animString += `75% { left:0px; top:40px; ${browser.cssPrefix}transform:rotate(270deg) scale(0.5)}`;
                animString += `100% { left:0px; top:0px; ${browser.cssPrefix}transform:rotate(360deg)scale(1)} }\n`;
                style.innerHTML += animString;

                ZCloader.zBlocks = true;
            }

            loadSetStyle(animBlock, {
                display: 'inline-block',
                verticalAlign: 'middle',
                width: '70px',
                height: '70px',
                transform: `scale(${baseObj.scale})`
            });

            var block1 = document.createElement('div');
            block1.innerHTML = '&nbsp';
            animBlock.appendChild(block1);
            loadSetStyle(block1, {
                backgroundColor: clr1,
                width: '30px',
                height: '30px',
                borderRadius: '1px',
                position: 'absolute',
                top: '0',
                left: '0',
                animationName: baseObj.type,
                animationDuration: `${baseObj.duration}s`,
                animationIterationCount: 'infinite',
                animationTimingFunction: 'ease-in-out'
            });

            var block2 = document.createElement('div');
            block2.innerHTML = '&nbsp';
            animBlock.appendChild(block2);
            loadSetStyle(block2, {
                backgroundColor: clr2,
                width: '30px',
                height: '30px',
                borderRadius: '1px',
                position: 'absolute',
                top: '0',
                left: '0',
                animationName: baseObj.type,
                animationDuration: `${baseObj.duration}s`,
                animationIterationCount: 'infinite',
                animationTimingFunction: 'ease-in-out',
                animationDelay: `-${(baseObj.duration / 4) * 1}s`
            });

            var block3 = document.createElement('div');
            block3.innerHTML = '&nbsp';
            animBlock.appendChild(block3);
            loadSetStyle(block3, {
                backgroundColor: clr3,
                width: '30px',
                height: '30px',
                borderRadius: '1px',
                position: 'absolute',
                top: '0',
                left: '0',
                animationName: baseObj.type,
                animationDuration: `${baseObj.duration}s`,
                animationIterationCount: 'infinite',
                animationTimingFunction: 'ease-in-out',
                animationDelay: `-${(baseObj.duration / 4) * 2}s`
            });

            var block4 = document.createElement('div');
            block4.innerHTML = '&nbsp';
            animBlock.appendChild(block4);
            loadSetStyle(block4, {
                backgroundColor: clr4,
                width: '30px',
                height: '30px',
                borderRadius: '1px',
                position: 'absolute',
                top: '0',
                left: '0',
                animationName: baseObj.type,
                animationDuration: `${baseObj.duration}s`,
                animationIterationCount: 'infinite',
                animationTimingFunction: 'ease-in-out',
                animationDelay: `-${(baseObj.duration / 4) * 3}s`
            });

            elemObj.loader = true;
        } else if (baseObj.type == 'zCircle1') {
            if (!ZCloader.zCircle1) {
                var animString = `\n@${browser.cssPrefix}keyframes ${baseObj.type} { `;
                animString += `0% ,25%, 50%, 75%, 100%{ ${browser.cssPrefix}transform: scale(0.0);  } `;
                animString += `12.5%  { ${browser.cssPrefix}transform: scale(1.0); background-color: ${clr1};}`;
                animString += `37.5%  { ${browser.cssPrefix}transform: scale(1.0); background-color: ${clr2};}`;
                animString += `62.5%  { ${browser.cssPrefix}transform: scale(1.0); background-color: ${clr3};}`;
                animString += `87.5%  { ${browser.cssPrefix}transform: scale(1.0); background-color: ${clr4};} }\n`;
                style.innerHTML += animString;

                ZCloader.zCircle1 = true;
            }

            loadSetStyle(animBlock, {
                display: 'inline-block',
                verticalAlign: 'middle',
                width: '70px',
                height: '70px',
                transform: `scale(${baseObj.scale})`
            });

            var circle = document.createElement('div');
            circle.innerHTML = '&nbsp';
            animBlock.appendChild(circle);
            loadSetStyle(circle, {
                width: '100%',
                height: '100%',
                borderRadius: '50%',
                opacity: '0.6',
                position: 'absolute',
                top: '0',
                left: '0',
                animationName: baseObj.type,
                animationDuration: `${baseObj.duration}s`,
                animationIterationCount: 'infinite',
                animationTimingFunction: 'ease-in-out'
            });

            elemObj.loader = true;
        } else if (baseObj.type == 'zBars') {
            if (!ZCloader.zBars) {
                var animString = `\n@${browser.cssPrefix}keyframes ${baseObj.type} { `;
                animString += `0%, 40%, 100% { ${browser.cssPrefix}transform: scaleY(0.4) }`;
                animString += `20% { ${browser.cssPrefix}transform: scaleY(1.0) } }\n`;
                style.innerHTML += animString;

                ZCloader.zBars = true;
            }

            loadSetStyle(animBlock, {
                display: 'inline-block',
                verticalAlign: 'middle',
                width: '50px',
                height: '30px',
                textAlign: 'center',
                fontSize: '10px',
                transform: `scale(${baseObj.scale})`
            });

            var bar1 = document.createElement('div');
            bar1.innerHTML = '&nbsp';
            animBlock.appendChild(bar1);
            loadSetStyle(bar1, {
                backgroundColor: clr1,
                marginLeft: '1px',
                display: 'inline-block',
                width: '6px',
                height: '100%',
                borderRadius: '1px',
                animationName: baseObj.type,
                animationDuration: `${baseObj.duration}s`,
                animationIterationCount: 'infinite',
                animationTimingFunction: 'ease-in-out'
            });

            var bar2 = document.createElement('div');
            bar2.innerHTML = '&nbsp';
            animBlock.appendChild(bar2);
            loadSetStyle(bar2, {
                backgroundColor: clr2,
                marginLeft: '1px',
                display: 'inline-block',
                width: '6px',
                height: '100%',
                borderRadius: '1px',
                animationName: baseObj.type,
                animationDuration: `${baseObj.duration}s`,
                animationIterationCount: 'infinite',
                animationTimingFunction: 'ease-in-out',
                animationDelay: `-${baseObj.duration - 0.1}s`
            });

            var bar3 = document.createElement('div');
            bar3.innerHTML = '&nbsp';
            animBlock.appendChild(bar3);
            loadSetStyle(bar3, {
                backgroundColor: clr3,
                marginLeft: '1px',
                display: 'inline-block',
                width: '6px',
                height: '100%',
                borderRadius: '1px',
                animationName: baseObj.type,
                animationDuration: `${baseObj.duration}s`,
                animationIterationCount: 'infinite',
                animationTimingFunction: 'ease-in-out',
                animationDelay: `-${baseObj.duration - 0.2}s`
            });

            var bar4 = document.createElement('div');
            bar4.innerHTML = '&nbsp';
            animBlock.appendChild(bar4);
            loadSetStyle(bar4, {
                backgroundColor: clr4,
                marginLeft: '1px',
                display: 'inline-block',
                width: '6px',
                height: '100%',
                borderRadius: '1px',
                animationName: baseObj.type,
                animationDuration: `${baseObj.duration}s`,
                animationIterationCount: 'infinite',
                animationTimingFunction: 'ease-in-out',
                animationDelay: `-${baseObj.duration - 0.3}s`
            });

            elemObj.loader = true;
        } else if (baseObj.type == 'zCircle2') {
            if (!ZCloader.zCircle2) {
                var animString = `\n@${browser.cssPrefix}keyframes ${baseObj.type} { `;
                animString += `0%, 80%, 100% { ${browser.cssPrefix}transform: scale(0.0) }`;
                animString += `40% { ${browser.cssPrefix}transform: scale(1.0) } }\n`;
                style.innerHTML += animString;

                ZCloader.zCircle2 = true;
            }

            loadSetStyle(animBlock, {
                display: 'inline-block',
                verticalAlign: 'middle',
                width: '72px',
                height: '18px',
                textAlign: 'center',
                fontSize: '10px',
                transform: `scale(${baseObj.scale})`
            });

            var circ1 = document.createElement('div');
            circ1.innerHTML = '&nbsp';
            animBlock.appendChild(circ1);
            loadSetStyle(circ1, {
                backgroundColor: clr1,
                display: 'inline-block',
                width: '18px',
                height: '18px',
                borderRadius: '100%',
                animationName: baseObj.type,
                animationDuration: `${baseObj.duration}s`,
                animationIterationCount: 'infinite',
                animationTimingFunction: 'ease-in-out'
            });

            var circ2 = document.createElement('div');
            circ2.innerHTML = '&nbsp';
            animBlock.appendChild(circ2);
            loadSetStyle(circ2, {
                backgroundColor: clr2,
                display: 'inline-block',
                width: '18px',
                height: '18px',
                borderRadius: '100%',
                animationName: baseObj.type,
                animationDuration: `${baseObj.duration}s`,
                animationIterationCount: 'infinite',
                animationTimingFunction: 'ease-in-out',
                animationDelay: `-${baseObj.duration - 0.2}s`
            });

            var circ3 = document.createElement('div');
            circ3.innerHTML = '&nbsp';
            animBlock.appendChild(circ3);
            loadSetStyle(circ3, {
                backgroundColor: clr3,
                display: 'inline-block',
                width: '18px',
                height: '18px',
                borderRadius: '100%',
                animationName: baseObj.type,
                animationDuration: `${baseObj.duration}s`,
                animationIterationCount: 'infinite',
                animationTimingFunction: 'ease-in-out',
                animationDelay: `-${baseObj.duration - 0.4}s`
            });

            var circ4 = document.createElement('div');
            circ4.innerHTML = '&nbsp';
            animBlock.appendChild(circ4);
            loadSetStyle(circ4, {
                backgroundColor: clr4,
                display: 'inline-block',
                width: '18px',
                height: '18px',
                borderRadius: '100%',
                animationName: baseObj.type,
                animationDuration: `${baseObj.duration}s`,
                animationIterationCount: 'infinite',
                animationTimingFunction: 'ease-in-out',
                animationDelay: `-${baseObj.duration - 0.6}s`
            });

            elemObj.loader = true;
        } else if (baseObj.type == 'zCrossBars') {
            if (!ZCloader.zCrossBars) {
                var animString = `\n@${browser.cssPrefix}keyframes ${baseObj.type} { `;
                animString += '0%{left:-50%;}';
                animString += '100%{left: 150%;} }\n';
                style.innerHTML += animString;

                ZCloader.zCrossBars = true;
            }

            loadSetStyle(animBlock, {
                display: 'inline-block',
                position: 'relative',
                verticalAlign: 'middle',
                width: '70px',
                height: '30px',
                boxSizing: 'border-box',
                borderWidth: '2px',
                borderStyle: 'solid',
                borderColor: 'rgb(160, 150, 150)',
                borderRadius: '8px',
                overflow: 'hidden',
                transform: `scale(${baseObj.scale})`
            });

            var bar1 = document.createElement('div');
            bar1.innerHTML = '&nbsp';
            animBlock.appendChild(bar1);
            loadSetStyle(bar1, {
                backgroundColor: clr1,
                width: '10px',
                height: '100%',
                position: 'absolute',
                transform: 'rotate(45deg) scaleY(2)',
                animationName: baseObj.type,
                animationDuration: `${baseObj.duration}s`,
                animationIterationCount: 'infinite',
                animationTimingFunction: 'linear',
                animationDelay: `-${baseObj.duration}s`
            });

            var bar2 = document.createElement('div');
            bar2.innerHTML = '&nbsp';
            animBlock.appendChild(bar2);
            loadSetStyle(bar2, {
                backgroundColor: clr2,
                width: '10px',
                height: '100%',
                position: 'absolute',
                transform: 'rotate(45deg) scaleY(2)',
                animationName: baseObj.type,
                animationDuration: `${baseObj.duration}s`,
                animationIterationCount: 'infinite',
                animationTimingFunction: 'linear',
                animationDelay: `-${(baseObj.duration / 100) * 80}s`
            });

            var bar3 = document.createElement('div');
            bar3.innerHTML = '&nbsp';
            animBlock.appendChild(bar3);
            loadSetStyle(bar3, {
                backgroundColor: clr3,
                width: '10px',
                height: '100%',
                position: 'absolute',
                transform: 'rotate(45deg) scaleY(2)',
                animationName: baseObj.type,
                animationDuration: `${baseObj.duration}s`,
                animationIterationCount: 'infinite',
                animationTimingFunction: 'linear',
                animationDelay: `-${(baseObj.duration / 100) * 60}s`
            });

            var bar4 = document.createElement('div');
            bar4.innerHTML = '&nbsp';
            animBlock.appendChild(bar4);
            loadSetStyle(bar4, {
                backgroundColor: clr4,
                width: '10px',
                height: '100%',
                position: 'absolute',
                transform: 'rotate(45deg) scaleY(2)',
                animationName: baseObj.type,
                animationDuration: `${baseObj.duration}s`,
                animationIterationCount: 'infinite',
                animationTimingFunction: 'linear',
                animationDelay: `-${(baseObj.duration / 100) * 40}s`
            });

            elemObj.loader = true;
        } else {
            loadSetStyle(animBlock, {
                display: 'inline-block',
                verticalAlign: 'middle',
                width: '30px',
                height: '30px',
                position: 'absolute',
                backgroundImage: `url('${_imageFilePath}loader_${baseObj.type}.gif')`,
                backgroundSize: '30px 30px',
                backgroundRepeat: 'no-repeat',
                backgroundPosition: 'center'
            });
        }

        // elemObj.stopLoader = function() {
        //     elemObj.removeChild(animBlock);
        //     elemObj.loader = false;
        //     delete elemObj.stopLoader;
        // }
        return animBlock;
    }
};
