// @zohocharts/zc/src/pie/piechart.js
//$Id$
//piechart.js
import { Registry } from '../core/zc-registry';
import { svg_plot_renderer, labelRenderer } from '../core/main-zc-init';
import {
    d3_pie,
    d3_arc,
    d3_select,
    d3_ease_linear,
    d3_scaleLinear,
    d3_lineRadial,
    d3_curves,
    d3_interpolate,
    d3_event
} from '../core/d3-constants';
import { polarToXY, getInnerRadius, polarRefPoint } from '../polaraxis/polarutils';
import { DIAL_CHARTNAME, PIE_CHARTNAME } from '../core/chartname-constants';
import { validateNullVal, utils, defined, mergeJSON, bbox, perToPx, pick, getValByPriority } from '../utils/zcutils';
import { mathMin, mathMax, mathAbs, mathPI } from '../core/zc-constants';
import { isDatalabelEnabled } from '../label/datalabelsUtils';
import { isBoundBoxEnabled, getBoundBoxPadding, getBoundBoxStrokeWidth } from '../label/boundBoxUtils';
import { getAnimationConf, cancelAnimation } from '../utils/animationutils';
import { getEventConf, getEventObject } from '../events/events-utils';
import { appendEle } from '../utils/zcsvgutils';
import { getFillColor, getStrokeColor, hasColorColumnInNonAxis } from '../utils/colorutils';
import { attachevents } from '../events/eventActions';
import { setPointerEvents } from '../events/events-utils';
import { PATH_ELEMENT, TEXT_ELEMENT } from '../core/element-constants';
import { generateDataForAnnotation } from '../annotation/annotation-utils';
import { FANONEBYONE_ANIMATION, FANALL_ANIMATION, INNER_ANIMATION } from '../core/animation-constants';
import { effects } from '../effects/zceffects';
import { INNERLABEL_LABELTYPE } from '../label/labeltype-constants';
import { arcInnerTween } from './pieAnimationUtils';
import { POINTER, NONE, NULL } from '../core/zc-constants';
import { $ZCG } from '../core/zcglobal';
import { updateSelectionOnDatapoint } from '../events/event-update-selection';
import { isDataActive } from '../core/data-processor/is-data-active';
import { isFacet } from '../axis/hierarchy-axis-utils';
import { DataProcessor } from '../core/data-processor';
import { setOpacityForTickFilterElements, setPivotAttributes } from '../plot/svg/set-pivot-attr';
import { getChartTypeByIndex } from '../core/charttype-helpers';
import { zcMorph } from '../morphing/zc-morping';
import { getRawColor } from '../utils/colorutils';
import { getOffsetValue } from '../plot/morph/morphTools';
import { d3_line } from '../core/d3-constants';

let annotation = Registry.getComponent('annotation');
/**
 *
 * @returns {chart} pie renderer.
 */
svg_plot_renderer.pie = function () {
    var MAXWIDTH_DATALABELS = 0.3;
    var pie = d3_pie().sort(NULL),
        arc = d3_arc(),
        sliceOut = function (d) {
            return polarToXY(sliceOutHeight, (d.startAngle + d.endAngle) / 2);
        },
        isredraw = false,
        InsidelabelTextGroup,
        labelLinkGroup,
        labelTextGroup,
        innerBoundBoxGroup,
        boundBoxGroup,
        hasNegValue,
        hasPreviousNegValue,
        gElementsCollected,
        hasBoundBox,
        previousRadiusFactors,
        previousInnerRadius,
        previousOuterRadius,
        baseValue;
    var commonRendererProp,
        chartObj,
        outerRadius,
        innerRadius,
        totalEngagement,
        color,
        container,
        piegroup,
        chartheight,
        chartwidth,
        pieArcs,
        paneid;
    var oldData, arcLabel, commonPlotOptions, systemconf, notesArray, isDial;
    var rad, OR, strokeWid, noofNonZeroDatas, sliceOutHeight;
    var modelOptions, transX, transY, customPlotarea, plotarea, dataindex;
    let shiftInfo;

    function chart(selection) {
        selection.each(function (data) {
            //renderer config values
            var rp = commonRendererProp.params;
            (chartObj = rp.chartObj),
                (commonPlotOptions = rp.commonPlotOptions),
                (color = rp.color),
                (notesArray = rp.notesArray),
                (systemconf = chartObj.systemConf);
            var labelQueue = rp.labelQueue,
                yaxiscolumnorder = rp.yaxiscolumnorder,
                dataset = rp.dataset,
                index = rp.datasetIndex || 0, //For pie,dial char, rp.datasetIndex is undefined
                { bound, rendererIndex = index } = rp;

            var defsLocation = chartObj.defsLocation,
                currentRadiusFactors = `${commonPlotOptions.maxRadius} _ ${commonPlotOptions.outerPadding} _ ${commonPlotOptions.innerRadius}`;
            paneid = isFacet(chartObj)
                ? `${getValByPriority([data, data[index]], 'row')}_${getValByPriority([data, data[index]], 'column')}`
                : null;
            let chartName = getChartTypeByIndex(chartObj, paneid, index).name;

            isDial = chartName === DIAL_CHARTNAME;
            var isPie = chartName === PIE_CHARTNAME,
                isBubblePie = !isDial && !isPie,
                serData = isBubblePie || (isFacet(chartObj) && hasColorColumnInNonAxis(chartObj)) ? data : data[index];

            dataindex = isBubblePie ? validateNullVal(commonPlotOptions.dataindex, 3) : dataset.yidx(yaxiscolumnorder);

            if (isDial) {
                plotarea = { ...chartObj.plotarea_approx, ...(bound && bound.polar) };
            } else {
                plotarea = { ...bound, ...customPlotarea };
            }

            (chartwidth = plotarea.width),
                (chartheight = plotarea.height),
                (outerRadius = validateNullVal(plotarea.radius, mathMin(chartwidth, chartheight) / 2));
            gElementsCollected = false;

            var plotoptions = serData.plotoptions,
                dataLabels = commonPlotOptions.datalabels,
                gvp = utils.getValueOfPlotOptions(plotoptions, commonPlotOptions);

            var { mode = 'radial' } = commonPlotOptions,
                maxRadius = validateNullVal(gvp('maxRadius'), 250),
                levelMarkerConf = commonPlotOptions.levelMarker,
                padding = isPie ? +gvp('outerPadding') : 0,
                strokeWidth = gvp('strokeWidth'),
                visibility,
                cornerRadius = gvp('cornerRadius') || 0,
                padAngle = gvp('padAngle') || 0,
                finalDisplay = (d) => {
                    var hidden = d.data.disabled || d.startAngle === d.endAngle || !d.value; // if disabled or value = 0 | null
                    return hidden ? NONE : NULL;
                },
                initialDisplay = function (d) {
                    var prevDisplay = d3_select(this).style('display'),
                        curDisplay = finalDisplay(d);
                    return !isredraw ||
                        (prevDisplay === NONE && curDisplay === NONE) ||
                        !isDataActive(d, chartObj, index) ||
                        (mode !== 'radial' && (d.data.disabled || d.data.dummy))
                        ? curDisplay
                        : NULL;
                };

            if (isFacet(chartObj) && padding > outerRadius) {
                // ZC2770
                padding = 0;
            }

            baseValue = commonPlotOptions.baseValue !== undefined ? commonPlotOptions.baseValue : 0;

            var showlabel = chartObj.datalabels.moduleExists && isDatalabelEnabled(dataLabels, chartObj),
                datalabelPadding = showlabel
                    ? +(dataLabels.padding || 0) +
                      (isBoundBoxEnabled(dataLabels, chartObj)
                          ? (getBoundBoxPadding(dataLabels, chartObj) + getBoundBoxStrokeWidth(dataLabels, chartObj)) * 2
                          : 0)
                    : 0,
                extraOptions = {},
                datalabelsWidth = 0;
            hasBoundBox = chartObj.datalabels.moduleExists && isBoundBoxEnabled(dataLabels, chartObj);

            padding += isPie ? datalabelPadding : 0;
            rad = mathMin(maxRadius, outerRadius - padding);
            if (isPie && showlabel) {
                /*
                 *If chart width is lesser than height, datalabels are hidden as default padding is 30 and there is no space to show datalabels.
                 * So 30% of the width is allocated for datalabels and the default padding will be changed to 2
                 */
                var minDatalabelsWidth = chartwidth * MAXWIDTH_DATALABELS,
                    availableDatalabelsWidth = chartwidth - 2 * rad;
                if (availableDatalabelsWidth < minDatalabelsWidth) {
                    datalabelPadding = extraOptions.padding = 5;
                    datalabelsWidth = mathMax(minDatalabelsWidth - availableDatalabelsWidth, 0);
                    rad -= datalabelsWidth / 2;
                }
            }
            innerRadius = getInnerRadius(commonPlotOptions, rad);
            strokeWidth = validateNullVal(strokeWidth, 1);

            if (isDial && levelMarkerConf.enabled) {
                var innerPadding, outerPadding;

                if (defined(commonPlotOptions.bandWidth) || defined(commonPlotOptions.maxBandWidth)) {
                    var maxDialBandWidth = validateNullVal(commonPlotOptions.maxBandWidth, rad);
                    var dialBandWidth = mathMin(perToPx(commonPlotOptions.bandWidth, rad, rad - innerRadius), maxDialBandWidth);

                    (innerPadding = -(rad - innerRadius - dialBandWidth) / 2), (outerPadding = -(rad - innerRadius - dialBandWidth) / 2);

                    visibility = dialBandWidth === 0 ? 'hidden' : NULL;
                } else {
                    innerPadding = mathMin(0, perToPx(levelMarkerConf.innerPadding, outerRadius, 0));
                    outerPadding = mathMin(0, perToPx(levelMarkerConf.outerPadding, outerRadius, 0));
                }

                rad += outerPadding;
                innerRadius -= innerPadding;
            }

            rad -= 0.5 * strokeWidth;
            OR = mathMax(innerRadius, rad);
            modelOptions =
                (plotoptions && plotoptions.models
                    ? mergeJSON(commonPlotOptions.models, plotoptions.models, true)
                    : commonPlotOptions.models) || {};
            //Animation variables
            var animationOptions = getAnimationConf(commonPlotOptions, plotoptions, chartObj, chartName);
            var isAnimationEnable = animationOptions.enabled,
                transitionDuration = !isAnimationEnable ? 0 : isredraw ? 200 : animationOptions.duration,
                easingType = isredraw ? d3_ease_linear : animationOptions.easingType,
                animationType = animationOptions.type,
                transitionEnd = function () {
                    d3_select(this).style('display', finalDisplay);
                };
            var eventDisabled = validateNullVal(getEventConf(plotoptions, commonPlotOptions, chartObj, 'enabled'), true) == false;
            var cursor = eventDisabled ? NULL : getEventConf(plotoptions, commonPlotOptions, chartObj, 'cursor') || POINTER;

            container = d3_select(this);
            piegroup = appendEle(container, 'g', [data], '.', 'class', 'piegroup');
            piegroup.styles({
                'stroke-width': strokeWidth,
                'fill-opacity': gvp('fillOpacity'),
                'stroke-opacity': commonPlotOptions.strokeOpacity,
                cursor: cursor,
                visibility: visibility
            });

            pie.startAngle(commonPlotOptions.startAngle)
                .endAngle(commonPlotOptions.endAngle)
                .value(function (d) {
                    return d.disabled ? 0 : mathAbs(d[dataindex] || 0);
                }); //d[dataindex] || 0 --> Why || 0 is, it may contain undefined Y value
            var pieData = pie((data = isBubblePie || (isFacet(chartObj) && hasColorColumnInNonAxis(chartObj)) ? data : serData.data[0]));
            chart.data = pieData;

            noofNonZeroDatas = pieData.reduce(function (prev, cur, i) {
                return prev + (cur.value ? 1 : 0);
            }, 0);
            chart.total = totalEngagement = pieData.reduce(function (prev, cur, i) {
                return prev + cur.value;
            }, 0);
            chart.arc = arc;

            if (mode === 'radial') {
                arc.outerRadius(OR).innerRadius(innerRadius);
            } else {
                let interPadding = perToPx(commonPlotOptions.interPadding, OR, 10),
                    len = data.reduce((prev, cur) => prev + (cur.disabled || cur.dummy ? 0 : 1), 0),
                    band = (OR - innerRadius - (len - 1) * interPadding) / len,
                    _index = -1;

                pieData.forEach((d) => {
                    if (!d.data.disabled && !d.data.dummy) {
                        _index++;
                    }
                    d.innerRadius = _index * (band + interPadding) + innerRadius;
                    d.outerRadius = (_index + 1) * band + _index * interPadding + innerRadius;
                });

                arc.startAngle(commonPlotOptions.startAngle);
            }
            arcLabel = d3_arc()
                .innerRadius(OR)
                .outerRadius(OR + datalabelPadding);
            var arcLabel1 = d3_arc().innerRadius(OR).outerRadius(OR);

            // translating the piegroup to center
            strokeWid = strokeWidth;
            var bb = translateWrap(piegroup);
            //calculating new radius
            var heightPercent = (bb.height + strokeWidth) / rad,
                widthPercent = (bb.width + strokeWidth) / rad;
            var newrad =
                rad +
                mathMin(
                    (chartwidth - 2 * padding - datalabelsWidth - (bb.width + strokeWidth)) / widthPercent,
                    (chartheight - 2 * padding - datalabelsWidth - (bb.height + strokeWidth)) / heightPercent
                );
            newrad = mathMin(maxRadius, mathMax(rad, newrad));
            if (isDial) {
                currentRadiusFactors += ` _ ${commonPlotOptions.bandWidth} _ ${levelMarkerConf.bandWidth} _ ${levelMarkerConf.innerPadding} _ ${levelMarkerConf.outerPadding}`;

                var yscale = rp.yscale || chartObj.axes.y.groups[0].scale;
                var minMax = yscale.domain();
                // var prev = yscale.domain()[0];
                pieData.forEach(function (d) {
                    // d.startAngle = yscale(prev);
                    // d.endAngle = yscale(prev + d.value);
                    // prev += d.value;
                    var val = d.data[dataindex];
                    hasPreviousNegValue = hasNegValue;
                    if ((hasNegValue = val < 0)) {
                        // Test_529.json
                        d.startAngle = yscale(baseValue != NULL ? Math.min(baseValue, minMax[1]) : minMax[1]);
                        d.endAngle = yscale(Math.max(val, minMax[0]));
                    } else {
                        d.startAngle = yscale(baseValue != NULL ? Math.max(baseValue, minMax[0]) : minMax[0]);
                        d.endAngle = yscale(Math.min(d.value, minMax[1]));
                    }
                });
            } else if (newrad != rad) {
                // updating values based on new radius
                rad = newrad;
                innerRadius = getInnerRadius(commonPlotOptions, rad);
                OR = mathMax(innerRadius, rad);
                arc.outerRadius(OR).innerRadius(innerRadius);
                arcLabel = d3_arc()
                    .innerRadius(OR)
                    .outerRadius(OR + datalabelPadding);
                arcLabel1 = d3_arc().innerRadius(OR).outerRadius(OR);
                translateWrap(piegroup);
            }
            // set the cornerRadius and padAngle
            arc.cornerRadius(cornerRadius).padAngle(padAngle);

            var modelID = !isDial && modelOptions.type ? `url(${defsLocation}#${createModelClippath(getModelPath())})` : NULL;
            var gradientDim = [OR, OR];
            var fillColor = (chart.fillColorFunc = getFillColor(chartObj, index, paneid, gradientDim)),
                strokeColor = getStrokeColor(chartObj, index, paneid);
            sliceOutHeight = perToPx(gvp('sliceOutHeight'), rad, 0) * 2;

            oldData = piegroup.selectAll(PATH_ELEMENT).data();

            pieArcs = appendEle(piegroup, PATH_ELEMENT, pieData);
            cancelAnimation(pieArcs);

            pieArcs
                .attrs({
                    'clip-path': modelID,
                    class: 'zc_pie',
                    id: function (d, i) {
                        return `arc_${i}`;
                    }
                })
                .styles({
                    display: initialDisplay,
                    fill: function (d, i) {
                        if (isFacet(chartObj)) {
                            return getFillColor(chartObj, i, paneid, gradientDim)(i, d.data);
                        }
                        return fillColor(i, d.data, null);
                    },
                    stroke: function (d, i) {
                        if (isFacet(chartObj)) {
                            return getStrokeColor(chartObj, i, paneid)(i, d.data, commonPlotOptions.strokeColor);
                        }

                        return strokeColor(i, d.data, commonPlotOptions.strokeColor);
                    }
                });

            setPivotAttributes(pieArcs, chartObj, 'pie');
            setOpacityForTickFilterElements(pieArcs, chartObj, 'pie');

            setPointerEvents(chartObj, pieArcs, eventDisabled, paneid);
            if (!chartObj.cache.search) {
                chartObj.cache.search = {};
            }
            const hasChartEvents = !Registry.getComponent('ZC_ChartEvents')._empty;
            if (!eventDisabled && hasChartEvents) {
                var mousemove = function (e, d) {
                        var args = getEventArgs(e, d);
                        chartObj.eventHandler.handlers.mousemove.call(chartObj.eventHandler, e, args);
                        updateSelectionOnDatapoint(d, args);
                        chartObj.cache.search.lastPane = paneid;
                    },
                    mouseout = function (e, d) {
                        var args = getEventArgs(e, d);
                        chartObj.eventHandler.handlers.mouseout.call(chartObj.eventHandler, e, args);
                        updateSelectionOnDatapoint(d, args);
                        chartObj.cache.search.lastPane = null;
                    },
                    click = function (e, d) {
                        var args = getEventArgs(e, d);
                        chartObj.eventHandler.handlers.click.call(chartObj.eventHandler, e, args); //chart,d swapped
                        updateSelectionOnDatapoint(d, args);
                        chartObj.cache.search.lastPane = paneid;
                    },
                    dblclick = function (e, d) {
                        var data = !annotation._empty ? generateDataForAnnotation(d.data, NULL) : d.data;
                        var coordinates = arcLabel.centroid(d);
                        coordinates = [coordinates[0] + transX, coordinates[1] + transY];
                        var args = getEventArgs(e, d);
                        args.coordinate = coordinates;
                        args.basedOn = 'plot';
                        args.data = data;
                        chartObj.eventHandler.handlers.doubleclick.call(chartObj.eventHandler, e, args);
                        updateSelectionOnDatapoint(d, args);
                        chartObj.cache.search.lastPane = paneid;
                    },
                    eventAttacherOptions = {
                        target: 'plot',
                        touchleave: true,
                        chartObj: chartObj
                    };
                var mouseHandlers = {
                        mousemove: mousemove,
                        mouseout: mouseout,
                        click: click,
                        dblclick: dblclick
                    },
                    touchHandlers = {
                        singletouchmove: mousemove,
                        touchend: mouseout,
                        tap: click,
                        dbltap: dblclick
                    };
                attachevents(pieArcs, mouseHandlers, touchHandlers, eventAttacherOptions);
            }
            if (/(\D)+onebyone/.test(animationType) && !isredraw) {
                var isFan = animationType == FANONEBYONE_ANIMATION;
                var duration = isFan ? transitionDuration : transitionDuration / noofNonZeroDatas;
                var prev = 0;
                pieArcs
                    .transition()
                    .ease(easingType)
                    .duration(function (d) {
                        return isFan ? duration * (d.value / totalEngagement) : duration;
                    })
                    .delay(function (d) {
                        var dur = duration * prev;
                        prev += isFan ? d.value / totalEngagement : 1;
                        return dur;
                    })
                    .call(isFan ? arcTween : innerTween)
                    .on('end', transitionEnd);
            } else {
                var tweenType;
                if (isredraw) {
                    tweenType = currentRadiusFactors != previousRadiusFactors ? innerTween : fanAllTween;
                } else {
                    tweenType = animationType == FANALL_ANIMATION ? fanAllTween : animationType == INNER_ANIMATION ? innerTween : arcTween;
                }

                if (chartObj.systemConf.chart.plot.morph.enabled && chartObj.oldSeriesInfo != undefined) {
                    var morph_Obj = new zcMorph({ sampling: chartObj.systemConf.chart.plot.morph.samplingDistance });
                    var oldseriesData = chartObj.oldSeriesInfo.get(index);
                    var oldseriesIndex = oldseriesData.seriesIndex;
                    var oldseriesColor = oldseriesData.color;
                    var offsetValue = getOffsetValue(chartObj, oldseriesIndex);
                    var individualElementTransform = oldseriesData.transformationInfo;
                    transitionDuration = chartObj.systemConf.chart.plot.morph.duration;
                    var line = d3_line();
                    var interpolated = {};

                    pieArcs.each(function (d, i) {
                        var fromPath = chartObj.oldSeriesInfo.get(oldseriesIndex).pathData[i];
                        var fromPathAttribute = fromPath.getAttribute('d');
                        fromPath.setAttribute('d', arc(d));
                        var toPath = fromPath.cloneNode();
                        fromPath.setAttribute('d', fromPathAttribute);
                        interpolated[i] = morph_Obj.interpolate(
                            {
                                path: fromPath,
                                xOffset: offsetValue.x + individualElementTransform[i].x,
                                yOffset: offsetValue.y + individualElementTransform[i].y
                            },
                            { path: toPath, xOffset: 0, yOffset: 0 }
                        );
                    });
                    pieArcs
                        .style('fill', function (d, i) {
                            return oldseriesColor[i];
                        })
                        .transition()
                        .duration(transitionDuration)
                        .ease(easingType)
                        .style('fill', function (d, i) {
                            return getRawColor(chartObj, i, d.data, oldseriesIndex, null);
                        })
                        .attrTween('d', function (d, i) {
                            var interpolator = interpolated[i];
                            return function (t) {
                                return line(
                                    interpolator.map(function (p) {
                                        return p(t);
                                    })
                                );
                            };
                        })
                        .on('end', function (d, i) {
                            if (chartObj.oldSeriesInfo) {
                                delete chartObj.oldSeriesInfo;
                            }
                            d3_select(this).style('fill', function () {
                                return fillColor(i, d.data, null);
                            });
                            this.setAttribute('d', arc(d));
                        });
                } else {
                    pieArcs.transition().duration(transitionDuration).ease(easingType).call(tweenType).on('end', transitionEnd);
                }
            }
            chartObj.timerObj[chartObj.timerObj.length] = chart.renderedTimer = setTimeout(function() {
                previousRadiusFactors = currentRadiusFactors;
                previousInnerRadius = innerRadius;
                previousOuterRadius = OR;
                if (isDial) {
                    chartObj.renderer[rendererIndex].endFunctions();
                } else {
                    if (labelRenderer && !isBubblePie) {
                        var translation = hasColorColumnInNonAxis(chartObj)
                            ? [transX + plotarea.translate[0], transY + plotarea.translate[1]]
                            : [transX, transY];

                        var args = [
                            piegroup.datum(),
                            0,
                            pieData,
                            OR,
                            totalEngagement,
                            arcLabel,
                            arcLabel1,
                            arc,
                            translation,
                            extraOptions,
                            chart.subSeriesIndex || 0
                        ];
                        labelQueue.push({
                            renderer: chart,
                            charttype: PIE_CHARTNAME,
                            chartObj: chartObj,
                            rendererIndex,
                            paneid: pick(paneid, index),
                            args: args
                        });
                        labelQueue.push({
                            renderer: chart,
                            charttype: 'pieInnerLabels',
                            chartObj: chartObj,
                            rendererIndex: rendererIndex,
                            paneid: pick(paneid, index),
                            args: args
                        });
                    }
                }

                //Annotation preprocessing
                var notes = chartObj.systemConf.notes;
                if (defined(notes) && notes.enabled && !isBubblePie && !annotation._empty) {
                    //MODULARIZE
                    notesArray = annotation.preprocessAnnotations(
                        chartObj.seriesdata,
                        0,
                        function (dt, k, j) {
                            var c = arcLabel.centroid(pieData[j]);
                            return [c[0] + transX, c[1] + transY];
                        },
                        notesArray,
                        chartObj,
                        rp
                    );
                }
                chart.chartRendered = true;
                isredraw = true;

                if (hasColorColumnInNonAxis(chartObj)) {
                    chartObj.renderer[rendererIndex].endFunctions();
                }
            }, transitionDuration + 1000 / 60);

            function getEventArgs(e, d) {
                const i = d.index;
                var ev = getEventObject(e);
                return {
                    renderer: chart,
                    seriesIndex: hasColorColumnInNonAxis(chartObj) ? i : 0,
                    subSeriesIndex: 0,
                    itemIndex: i,
                    paneid: paneid,
                    rendererIndex: rendererIndex,
                    color: color[i % color.length],
                    coordinates: [ev.pageX, ev.pageY],
                    element: this,
                    data: d,
                    selected: d.selected,
                    _key: DataProcessor.helpers.generatePointKey(chartObj.id, paneid, 0, 0, i)
                };
            }
        });
        return chart;
    }

    chart.commonRendererProp = function (_) {
        if (!arguments.length) {
            return commonRendererProp;
        }
        commonRendererProp = _;
        return chart;
    };

    chart.plotarea = function (_) {
        //Bubble Pie chart will give you the plotarea
        if (!arguments.length) {
            return customPlotarea || plotarea;
        }
        customPlotarea = _;
        return chart;
    };

    chart.innerRadius = function (_) {
        if (!arguments.length) {
            return innerRadius;
        }
        innerRadius = _;
        return chart;
    };

    chart.defaultMouseOverFunc = function (d, seriesPos, itemPos) {
        if (noofNonZeroDatas == 1) {
            return;
        }

        if (!gElementsCollected) {
            collectGroups(commonRendererProp.params.rendererIndex);
        }

        var pt = sliceOut(d),
            subSeriesIndex = chart.subSeriesIndex || 0;
        var slicedPosition = {
            transform: `translate(${pt[0]},${pt[1]})`
        };

        pieArcs.transition().ease(d3_ease_linear).attr('transform', 'translate(0,0)');

        //Reset previous pane mouseover
        // if(chartObj.dataObject.pivot){
        //     var allPieGroups = chartObj.wrap.selectAll('g.piegroup'),
        //         allPieArcs = allPieGroups.selectAll('path');
        //     allPieArcs
        //         .transition()
        //         .ease(d3_ease_linear)
        //         .attr('transform', 'translate(0,0)');
        // }

        piegroup.selectAll(`#arc_${itemPos}`).transition().ease(d3_ease_linear).attrs(slicedPosition);

        InsidelabelTextGroup.selectAll(`#innerlbl0_${subSeriesIndex}_${itemPos}`).transition().ease(d3_ease_linear).attrs(slicedPosition);

        if (!hasBoundBox) {
            labelLinkGroup.selectAll(`#link0_${subSeriesIndex}_${itemPos}`).style('stroke-width', 3 * chart.lineStrokeWidth);
        }

        var label = labelTextGroup.selectAll(`#lbl0_${subSeriesIndex}_${itemPos}`).node(),
            boundBoxSlicedPos = label
                ? {
                      transform: 'translate(' + (Number(label.getAttribute('x'))+pt[0]) + ',' + (Number(label.getAttribute('y'))+pt[1]) + ')'
                  }
                : slicedPosition,
            innerlabel = InsidelabelTextGroup.selectAll(`#innerlbl0_${subSeriesIndex}_${itemPos}`).node(),
            innerBoundBoxSlicedPos = innerlabel
                ? {
                      transform: 'translate(' + (Number(innerlabel.getAttribute('x'))+pt[0]) + ',' + (Number(innerlabel.getAttribute('y'))+pt[1]) + ')'
                  }
                : slicedPosition;
        
        if (chart.showlabel == 'circular') {
            labelTextGroup.selectAll(`#lbl0_${subSeriesIndex}_${itemPos}`).transition().ease(d3_ease_linear).attrs(slicedPosition);

            boundBoxGroup.selectAll(`#boundboxlbl0_${subSeriesIndex}_${itemPos}`).transition().ease(d3_ease_linear).attrs(boundBoxSlicedPos);
        }
        labelTextGroup.selectAll(`#lbl0_${subSeriesIndex}_${itemPos}`).style('text-shadow', '0 0 0 rgba(0,0,0,1)');

        innerBoundBoxGroup
            .selectAll(`#boundboxinnerlbl0_${subSeriesIndex}_${itemPos}`)
            .transition()
            .ease(d3_ease_linear)
            .attrs(innerBoundBoxSlicedPos);
    };

    chart.defaultMouseOutFunc = function (d, i, forceMouseOut) {
        if (!gElementsCollected) {
            collectGroups(commonRendererProp.params.rendererIndex);
        }

        var actualPosition = {
            transform: 'translate(0,0)'
        };

        labelTextGroup.selectAll(TEXT_ELEMENT).style('text-shadow', NULL);
        if (!hasBoundBox) {
            labelLinkGroup.selectAll(PATH_ELEMENT).style('stroke-width', chart.lineStrokeWidth);
        }

        var subSeriesIndex = chart.subSeriesIndex || 0,
            label = labelTextGroup.selectAll(`#lbl0_${subSeriesIndex}_${d.itemIndex}`).node(),
            boundBoxSlicedPos = label
                ? {
                      transform: 'translate(' + label.getAttribute('x') + ',' + label.getAttribute('y') + ')'
                  }
                : actualPosition,
            innerlabel = InsidelabelTextGroup.selectAll(`#innerlbl0_${subSeriesIndex}_${d.itemIndex}`).node(),
            innerBoundBoxSlicedPos = innerlabel
                ? {
                      transform: 'translate(' + innerlabel.getAttribute('x') + ',' + innerlabel.getAttribute('y') + ')'
                  }
                : actualPosition;

        if (!$ZCG.clickedTime && forceMouseOut) {
            pieArcs.attrs(actualPosition);

            InsidelabelTextGroup.selectAll(TEXT_ELEMENT).attrs(actualPosition);

            innerBoundBoxGroup.selectAll(`#boundboxinnerlbl0_${subSeriesIndex}_${d.itemIndex}`).transition().ease(d3_ease_linear).attrs(innerBoundBoxSlicedPos);

            if (chart.showlabel == 'circular') {
                labelTextGroup.selectAll(TEXT_ELEMENT).attrs(actualPosition);
                boundBoxGroup.selectAll(`#boundboxlbl0_${subSeriesIndex}_${d.itemIndex}`).transition().ease(d3_ease_linear).attrs(boundBoxSlicedPos);
            }
            return;
        }
        pieArcs.transition().ease(d3_ease_linear).attrs(actualPosition);

        InsidelabelTextGroup.selectAll(TEXT_ELEMENT).transition().ease(d3_ease_linear).attrs(actualPosition);

        if (chart.showlabel == 'circular') {
            labelTextGroup.selectAll(TEXT_ELEMENT).transition().ease(d3_ease_linear).attrs(actualPosition);
            boundBoxGroup.selectAll(`#boundboxlbl0_${subSeriesIndex}_${d.itemIndex}`).transition().ease(d3_ease_linear).attrs(boundBoxSlicedPos);
        }
        innerBoundBoxGroup.selectAll(`#boundboxinnerlbl0_${subSeriesIndex}_${d.itemIndex}`).transition().ease(d3_ease_linear).attrs(innerBoundBoxSlicedPos);
    };

    function getGearData(count) {
        var array = [],
            data = [],
            x = 0;
        for (var i = 0; i < count; i++) {
            modelOptions.type == 'gear1' ? data.push([x++, 0], [x++, 1], [x++, 1]) : data.push([x, 0], [x++, 1], [x, 1], [x++, 0]);
        }
        array.push(data);
        return array;
    }

    function getModelPath() {
        var index = -1,
            count = modelOptions.count,
            min = modelOptions.type == 'zigzag' ? -50 + innerRadius / 5 : modelOptions.type == 'sun' ? -1 : -10;
        var gearData = getGearData(count);
        var angleScale = d3_scaleLinear()
            .range([0, 2 * mathPI])
            .domain([0, modelOptions.type == 'gear1' ? 3 * count : 2 * count]);
        var radiusScale = d3_scaleLinear().range([innerRadius, OR]).domain([min, 1]);
        var radialLine = d3_lineRadial()
            .curve(d3_curves('linearClosed'))
            .radius(function (d) {
                return radiusScale(d[1]);
            })
            .angle(function (d, i) {
                index =
                    modelOptions.type == 'gear'
                        ? i % 2 == 0
                            ? index + 1
                            : index
                        : modelOptions.type != 'gear1'
                        ? i % 2 != 0
                            ? index + 1
                            : index
                        : index;
                return angleScale(modelOptions.type == 'gear1' ? index++ : index);
            });
        return radialLine(gearData[0]);
    }

    function createModelClippath(gearPath) {
        var properties = {
                name: 'pieModels',
                id: chartObj.id,
                d: gearPath
            },
            args = {
                chartObj: chartObj
            };
        var clipID = effects.get(effects.svg_clips, 'polygon', properties, args);
        return clipID;
    }

    function innerTween(element) {
        arcInnerTween(
            element,
            arc,
            [isredraw ? previousInnerRadius : innerRadius, innerRadius],
            [isredraw ? previousOuterRadius : innerRadius, OR]
        );
    }

    function arcTween(transition) {
        transition.attrTween('d', function (d, i) {
            var interpolate = d3_interpolate(d.startAngle, d.endAngle);
            return function (t) {
                d.endAngle = interpolate(t);
                return arc(d);
            };
        });
    }

    function fanAllTween(transition) {
        transition.attrTween('d', function (d, i) {
            var od = oldData[i] || oldData[oldData.length - 1],
                prevEA,
                prevSA;

            if (hasNegValue) {
                (prevEA = isredraw && od && hasPreviousNegValue ? od.endAngle : d.startAngle),
                    (prevSA = isredraw && od && hasPreviousNegValue ? od.startAngle : d.startAngle);
            } else {
                (prevEA = isredraw && od && !hasPreviousNegValue ? od.endAngle : isDial ? d.startAngle : commonPlotOptions.startAngle),
                    (prevSA =
                        isredraw && od && !hasPreviousNegValue
                            ? oldData[i]
                                ? od.startAngle
                                : od.endAngle
                            : isDial
                            ? d.startAngle
                            : commonPlotOptions.startAngle);
            }

            var interpolate = d3_interpolate(prevEA, d.endAngle);
            var interpolate1 = d3_interpolate(prevSA, d.startAngle);

            return function (t) {
                d.endAngle = interpolate(t) || 0;
                d.startAngle = interpolate1(t) || 0;
                return arc(d, i);
            };
        });
    }

    function translateWrap(wrap) {
        //dummy data
        var dummyData = [];
        dummyData[dataindex] = 1;
        let dummyPieData = pie([dummyData])[0];
        dummyPieData.innerRadius = 0;
        dummyPieData.outerRadius = outerRadius;

        //dummy pie for getting the bound
        var dummy = container
            .append(PATH_ELEMENT)
            .attr('d', arc(dummyPieData, chart.data.length - 1))
            .styles({
                fill: NONE,
                stroke: NONE
            });
        var bb = bbox(dummy.node());
        dummy.remove();

        //Reference point calculation for pie layout
        var pt = polarRefPoint(bb, chartwidth, chartheight, rad, strokeWid);
        (transX = isDial ? plotarea.x : pt[0]), (transY = isDial ? plotarea.y : pt[1]);
        (chartObj.plotarea.centerX = transX), (chartObj.plotarea.centerY = transY);
        wrap.attr('transform', function (d, i) {
            return `translate(${transX},${transY})`;
        });
        return bb;
    }

    function collectGroups(rendererIndex = 0) {
        let subSeriesIndex = chart.subSeriesIndex || 0,
            parentWrap = isFacet(chartObj) ? chartObj.staticPaneSVG.selectAll(`g.grid${paneid}`) : chartObj.staticPaneSVG;
        InsidelabelTextGroup =
            InsidelabelTextGroup || parentWrap.selectAll(`g.${INNERLABEL_LABELTYPE}_${rendererIndex} g.subseries_${subSeriesIndex}`);
        labelLinkGroup = labelLinkGroup || parentWrap.selectAll(`g.link_${rendererIndex} g.subseries_${subSeriesIndex}`);
        labelTextGroup = labelTextGroup || parentWrap.selectAll(`g.label_${rendererIndex} g.subseries_${subSeriesIndex}`);
        boundBoxGroup = boundBoxGroup || parentWrap.selectAll(`g.boundboxseries_${rendererIndex} g.subseries_${subSeriesIndex}`);
        innerBoundBoxGroup =
            innerBoundBoxGroup || parentWrap.selectAll(`g.innerboundboxseries_${rendererIndex} g.subseries_${subSeriesIndex}`);
        gElementsCollected = true;
    }

    chart.shiftInfo = function(_) {
        return arguments.length ? (shiftInfo = _, chart): shiftInfo;
    }

    chart.destroy = function destroy() {
        container = pieArcs = piegroup = null;
    }

    return chart;
};

export var _pieChart = svg_plot_renderer.pie;
Registry.setComponent('svgRenderer.pie', _pieChart);
