// @zohocharts/zc/src/pie/pieOptions.js
import { defaultTheme as options } from '../core/defaulttheme';
import { mathPI } from '../core/zc-constants';

options.pie = function(type) {
    var defaultOptions = {
        startAngle: 0,

        endAngle: 2 * mathPI,
        innerRadius: 0,
        outerPadding: 10,
        datalabels: {
            type: 'circular', // circular || doubleside
            padding: 5,
            line: {
                strokeColor: 'grey',
                strokeWidth: 0.3
            },
            innerLabel: {
                show: false,
                showAs: 'percent' //x,y,percent
            }
        },
        animation: {
            type: 'fanAll' //inner, wheel, fanAll, fanonebyone , inneronebyone
        },
        gradients: {
            // radial || linear || none
            options: {
                linear: {
                    x2: 15,
                    spreadMethod: 'repeat', // pad || repeat || reflect
                    gradientUnits: 'userSpaceOnUse' //objectBoundingBox || userSpaceOnUse
                },
                radial: {
                    radius: 90
                }
            }
        },
        models: {
            count: 50
        }
    };
    if (type === 'bubblepie' || type === 'geobubblepie') {
        defaultOptions.strokeColor = 'transparent';
        defaultOptions.outerPadding = 0.02;
        defaultOptions.datalabels.showAs = 'percent'; //x,y,percent
        defaultOptions.hoveroptions = {
            innerFillColor: 'transparent',
            innerStrokeColor: 'transparent',
            outerFillColor: 'transparent',
            outerStrokeWidth: 3
        };
    } else {
        defaultOptions.strokeColor = 'white';
        defaultOptions.sliceOutHeight = 5;
        defaultOptions.maxRadius = 250;
        defaultOptions.datalabels.showAs = 'x'; //x,y,percent
    }
    return defaultOptions;
};

export var _pieOptions = options.pie;
