// @zohocharts/zc/src/events/listeners/handler-mapping.js
import { ZC_PLOT_HOVER } from './hover';
import { ZC_PLOT_MOUSEOUT } from './mouseout';
import { ZC_PLOT_CLICK } from './click';
import { ZC_PLOT_DOUBLECLICK } from './doubleclick';
import { ZC_PLOT_CONTEXTMENU } from './contextmenu';

export function _defaultPlotHandlers() {
    return {
        mousemove: (event, args) => ZC_PLOT_HOVER.call(this, event, args),
        mouseout: (event, args) => ZC_PLOT_MOUSEOUT.call(this, event, args),
        click: (event, args) => ZC_PLOT_CLICK.call(this, event, args),
        doubleclick: (event, args) => ZC_PLOT_DOUBLECLICK.call(this, event, args),
        contextmenu: (event, args) => ZC_PLOT_CONTEXTMENU.call(this, event, args),
        tap: (event, args) => ZC_PLOT_CLICK.call(this, event, args),
        dbltap: (event, args) => ZC_PLOT_DOUBLECLICK.call(this, event, args),
        singletouchmove: (event, args) => ZC_PLOT_HOVER.call(this, event, args),
        touchend: (event, args) => ZC_PLOT_MOUSEOUT.call(this, event, args)
    };
}


export let __defaultPlotHandlers = {

    mousemove: ZC_PLOT_HOVER,
    mouseout: ZC_PLOT_MOUSEOUT,
    click: ZC_PLOT_CLICK,
    doubleclick: ZC_PLOT_DOUBLECLICK,
    contextmenu: ZC_PLOT_CONTEXTMENU,
    tap: ZC_PLOT_CLICK,
    dbltap: ZC_PLOT_DOUBLECLICK,
    singletouchmove: ZC_PLOT_HOVER,
    touchend: ZC_PLOT_MOUSEOUT
}
