// @zohocharts/zc/src/scale-parser/legend/size/size-domain-parser.js
//$Id$
import { NUMERIC_DATATYPE, TIME_DATATYPE } from '../../../core/datatype-constants';
import { UNDEFINED } from '../../../core/zc-constants';
import { d3_merge, d3_ascending } from '../../../core/d3-constants';
import { d3_values } from '../../../utils/zcutils';
import { ScaleRegistry } from '../../../scale/scale-registry';
import { DataProcessor } from '../../../core/data-processor';
import { fetchForSizdDomainParser } from '../../fetch-operators';
import { convertNull } from '../../../core/data-processor/size/convert-null';
import { categories } from '../../../scale/domain/categories';
import { toTimestamp } from '../../../scale/scale-component';

export function sizeDomainParser(encodeType, index, config, chart) {
    let processor = chart.dataObject,
        dataSet = chart.dataset,
        dataType = dataSet.getDataType(encodeType, index),
        columnInfo = dataSet.getColumn(encodeType, index),
        nullAsOptions = { config, columnKey: dataSet.getColumnIdx(encodeType, index) },
        zEncode = processor.encodes && processor.getEncode(encodeType, index),
        _data = config._data || (zEncode ? zEncode.seriesData : []),
        //
        isCatNull = DataProcessor.helpers.isCatNull,
        isCatNullX = isCatNull(dataSet, 'x'),
        isCatNullY = isCatNull(dataSet, 'y', 0),
        condition =
            hasGeo(chart) || (chart.systemConf.legend.singleLayout && dataType === NUMERIC_DATATYPE)
                ? false
                : d => !(isCatNullX(dataSet.getActualX(d, index)) || isCatNullY(dataSet.getY(d, index))),
        schema = dataType === TIME_DATATYPE ? toTimestamp(config, columnInfo) : null,
        { minRange, maxRange, stops, ranges } = schema || config,
        nullAsConvertion = d => convertNull(d[columnInfo.dataindex], nullAsOptions);

    // _data = _data.map(series => {
    //     series.data[0].filter(condition)
    // })

    // if (ranges && ranges._dummy) {
    //     config.ranges = null;
    //     ranges = null;
    // }

    // TODO
    // let view = _legendType(encodeType, { config, dataType }),
    //     newRanges = generateRanges(config, schema, scale, view);
    //
    // if (newRanges) {
    //     config.ranges = newRanges;
    // }

    // if (encodeType === 'z') {
    //     // getLegendConfig(encodeType, index, chart).ranges = ranges;
    // }

    let domainOperators = {},
        defaultOperators = fetchForSizdDomainParser(dataType);

    /*
        EXTREMES
     */
    if (defaultOperators.has('extremes')) {
        if (chart.systemConf.legend.singleLayout && encodeType === 'z') {
            domainOperators.extremes = [() => processor.getMinMaxZ()];
        } else {
            domainOperators.extremes = [
                {
                    _type: 'extremesByAccessor',
                    data: _data,
                    condition,
                    accessor: function() {
                        return nullAsConvertion;
                    }
                }
            ];
        }
    }

    if (defaultOperators.has('categories')) {
        let _d = d3_merge(_data.map(_ => d3_merge(_.data)));
        domainOperators.categories = categories(_d, nullAsConvertion);
    }

    if (dataType === NUMERIC_DATATYPE && !chart.isGeoMap) {
        let base = config.base;
        domainOperators.base = base !== UNDEFINED ? base : 0;
    }

    /*
        Custom min max range
     */
    domainOperators.min = minRange;
    domainOperators.max = maxRange;

    if (ranges && ranges.length) {
        /* Ranges*/
        domainOperators.sortedRanges = {
            datatype: dataType,
            clone: false,
            config: { ...config, ranges }
        };
    } else if (stops && stops.length) {
        /*Stops*/
        domainOperators.filteredStops = {
            array: d3_values(stops).sort(d3_ascending),
            min: minRange,
            max: maxRange
        };
    }
    return domainOperators;
}

// function generateRanges(config, schema, scale, view) {
//     let { dataType, meta } = schema.data,
//         { ranges, stops } = schema.domain,
//         configRanges = config.ranges,
//         subFunction = getSubFunction(meta),
//         scale_domain = scale.domain();
//
//     if (view !== LIST_LEGEND_TYPE || ranges || dataType === ORDINAL_DATATYPE) {
//         return;
//     }
//
//     // If no change in scale domain and subfunction, just return the ranges(reason: after legend filter if chart dimension differs on redraw, new set of ranges are created in where disabled property is removed)
//     let _cache = configRanges && configRanges._cache;
//     if (_cache && _cache.subFunction === subFunction && jsonStringify(_cache.domain) === jsonStringify(scale_domain)) {
//         return configRanges;
//     }
//
//     let count = getTickCount(config, scale, view),
//         ticks = stops ? scale_domain : getEntries(scale, dataType, count + 1, subFunction),
//         generated = ticks.length === 1 ? [[ticks[0], ticks[0]]] : d3_pairs(ticks); // ZC2548
//
//     generated._dummy = true;
//     generated._cache = { subFunction, domain: scale_domain };
//
//     // Adding the disabled property on redraw
//     if (configRanges && jsonStringify(configRanges) === jsonStringify(generated)) {
//         configRanges.forEach((_, i) => (generated[i].disabled = _.disabled));
//     }
//
//     return generated;
// }
//
// function getSubFunction(meta) {
//     let { datatype: dataType } = meta,
//         formatConfig = meta[dataType] || {};
//
//     return formatConfig.subfunction;
// }

ScaleRegistry.add('size.domainParser', sizeDomainParser);

function hasGeo(chartObj) {
    // TODO  change chartObj.dataObject.isFlowChartCategory
    return chartObj.isGeoMap || chartObj.mapObj || chartObj.dataObject.isFlowChartCategory;
}
