// @zohocharts/zc/src/scale-parser/legend/size/size-legend-parser.js
//$Id$
import { Registry } from '../../../core/zc-registry';
import { ScaleRegistry } from '../../../scale/scale-registry';

export function sizeLegendParser(encodeType, index = 0, chart, config) {
    let dataSet = chart.dataset,
        dataType = dataSet.getDataType(encodeType, index),
        columnInfo = dataSet.getColumn(encodeType, index);

    var parser = {
        type: dataType,
        dataType: dataType,
        scaleType: columnInfo.scaleType,
        field: columnInfo,
        reverse: config.reversed,
        clamp: true,
        data: [],
        domain: {
            operators: ScaleRegistry.get('size.domainParser')(encodeType, index, config, chart),
            evaluator: ScaleRegistry.get('size.evalDomain')
        },
        range: {
            operators: ScaleRegistry.get('size.rangeParser')(encodeType, index, config, chart),
            evaluator: ScaleRegistry.get('size.evalRange')
        }
    };

    return parser;
}

Registry.setComponent('size.parser', sizeLegendParser);
