// @zohocharts/zc/src/parser/parser-flow.js
import { Registry } from '../core/zc-registry';
import { invokeFunction, isArray, defined, getPropVal, setPropVal } from '../utils/zcutils';
import { d3_set, d3_merge } from '../core/d3-constants';
export function ViewFlow(spec, oldSpec, bound, oldInstance) {
    this.spec = spec;
    this.data = spec.data ||  (oldSpec.seriesdata && oldSpec.seriesdata.chartdata);
    this.oldInstance = oldInstance;
    this.encoding = {};
    this.columns = this.spec.columns || oldSpec.metadata.columns;
    this.columnInfo = {};
    this.processedData = {};
    this.scales = {};
    this.axes = {
        vertical: new Array(0),
        horizontal: new Array(0)
    };
    this.marks = {
        _default: {},
        _custom: new Array(0)
    };
    this.marks._default = {
        chart: new Array(0),
        label: new Array(0)
    };
    this.legends = {};
    this.reacts = {};
    this.triggerQueue = [];
    this.bound = bound;
    this.transition = {
        start: false,
        end: false,
        count: 0
    }
}

ViewFlow.prototype = {
    init: function() {
        if (this.spec.encoding) {
            this.addEncode(this.spec.encoding);
        }
        if(this.spec.transform) {
            for(let key in this.spec.transform){
                if(key == 'layout'){
                    let layouts = this.spec.transform[key];
                    layouts.map(layoutConfig => {
                        if(layoutConfig.type == 'tree'){
                            this.hierarchy = true;
                        }
                    })
                }
            }
        }
    },
    process: function(_data){
        if(this.spec.transform){
            for(let key in this.spec.transform){
                if(key == 'layout'){
                    let layouts = this.spec.transform[key];
                    layouts.map(layoutConfig => {
                        if(layoutConfig.scale){
                            this.scales[layoutConfig.scale].dependencies.push({
                                "name": 'layout',
                                "id": layoutConfig.id
                            })
                            this.scales[layoutConfig.scale].dependencyIds.add(layoutConfig.id);
                        }

                        this.applyTransform(key, layoutConfig, _data);
                    })
                    
                }else if(key == 'groupBy'){
                    let groupBys = this.spec.transform[key];
                    groupBys.map(groupByConfig => {

                        this.applyTransform(key, groupByConfig, _data);
                    })
                }
            }
        }
        if(this.spec.reacts) {
            this.spec.reacts.map(react => this.addReacts(react));
        }
        if (this.spec.scales) {
            this.spec.scales.map(scale => this.addScale(scale));
        }
        if (this.spec.axes) {
            this.spec.axes.map(axis => this.addAxis(axis));
        }
        if (this.spec.marks) {
            this.spec.marks.map(mark => this.addMarks(mark));
        }
        if(this.spec.legends) {
            this.spec.legends.map(mark => this.addLegend(mark));
        }
    },
    addLegend: function(_legend){
        this.legends[_legend.id] = {
            conf: _legend
        };

        if(_legend.scale && this.scales[_legend.scale]){
            this.scales[_legend.scale].dependencies.push({
                "name": 'legend',
                "id": _legend.id
            })

            this.scales[_legend.scale].dependencyIds.add(_legend.id);
        }
        
    },
    addEncode: function(_encodes) {
        //encode x
        if (_encodes.x) {
            this.encoding.x = [];
            if (_encodes.x.constructor == Array) {
                _encodes.x.map(d => {
                    this.encoding.x.push({
                        column: d
                    });
                });
            } else {
                _encodes.x.map(d => {
                    this.encoding.x.push({
                        column: d
                    });
                });
            }
        }

        //encode y
        if (_encodes.y) {
            this.encoding.y = [];
            _encodes.y.map(d => {
                this.encoding.x.push({
                    column: d
                });
            });
        }
    },
    applyTransform: function(key, transformSpec, _data){
        let transformed_data = Registry.getComponent(transformSpec.type+'_'+key).call(this, transformSpec, _data);

        //For saving object internally inside the viewparser
        if(transformSpec.as){
            let needToSaveOutput = transformSpec.as.output;

            if( needToSaveOutput ){
                this.processedData[transformSpec.as.output] = transformed_data;
            }
        }

        //For exporting saved data to outside parser object like chartObj instances
        if(transformSpec.export){
            for(let key in transformSpec.export){
                let exportToVariable = transformSpec.export[key],
                fromObj = getPropVal(transformed_data, key.replace("output.", ""));
                setPropVal(this, exportToVariable.replace("this.", ""), fromObj);
            }
        }
    },
    addScale: function(_scale) {
        this.scales[_scale.id] = {
            conf: _scale,
            dependencies: new Array(0),
            dependencyIds: d3_set()
        };
        this.mapColumnToScale(_scale);
    },
    addAxis: function(_axis) {
        if (_axis.orient == 'left' || _axis.orient == 'right') {
            this.axes.vertical.push({
                conf: _axis,
                scale: this.scales[_axis.scale]
            });
        } else {
            this.axes.horizontal.push({
                conf: _axis,
                scale: this.scales[_axis.scale]
            });
        }
        if(_axis.scale && this.scales[_axis.scale]){
            this.scales[_axis.scale].dependencies.push({
                "type": 'axis',
                "id": _axis.id
            })
    
            this.scales[_axis.scale].dependencyIds.add(_axis.id);
        }
        
    },
    addMarks: function(_mark) {
        if (_mark.type == 'DEFAULT') {
            let _data = invokeFunction(_mark.data, this.data),
                renderer = _data.map((d, i) => invokeFunction(_mark.renderer, d, i)),
                markOperator,
                markSpec;

            if (_mark.operator) {
                markOperator = _data.map((d, i) => invokeFunction(_mark.operator, d, i));
            }

            if (_mark.spec) {
                markSpec = _data.map((d, i) => {
                    if (typeof _mark.spec == 'function') {
                        return _mark.spec(d, i);
                    } else {
                        return _mark.spec;
                    }
                });
            }
            _data.map((d, i) => {
                if (renderer[i].isDataLabel) {
                    this.marks._default.label.push({
                        data: d,
                        operator: markOperator && markOperator[i],
                        type: renderer[i].type,
                        spec: markSpec && markSpec[i]
                    });
                } else {
                    this.marks._default.chart.push({
                        data: d,
                        operator: markOperator && markOperator[i],
                        type: renderer[i].type,
                        datasetIndex: renderer[i].datasetIndex,
                        spec: markSpec && markSpec[i],
                        x: renderer[i].x && this.scales[renderer[i].x].component,
                        y: renderer[i].y && this.scales[renderer[i].y].component
                        // spec:
                    });
                }
            });
        } else {
            this.marks._custom.push({
                mark: _mark
            });
        }
    },
    addReacts: function(react){
        let _ = {},
            trigger = react.trigger;
        _.value = react.value;
        _.dependencies = new Array(0);
        _.dependencyIds = d3_set();

        this.reacts[react.id] = _;
        
        if(trigger){
            this.triggerQueue.push({
                trigger,
                react: _
            })
        }
        
        
    },
    mapColumnToScale: function(_scale){
        let columnName = _scale.column;
        if(this.columns){
            this.columns.filter(column => {
                if(column.columnname == columnName){
                    if(this.columnInfo[columnName]){
                        let existingScaleId = this.columnInfo[columnName].scale_id;
                        existingScaleId = isArray(existingScaleId) ? existingScaleId : [existingScaleId];
                        this.columnInfo[columnName] = {
                            scale_id: d3_merge([existingScaleId, [_scale.id]])
                        }
                    }else{
                        this.columnInfo[columnName] = {
                            scale_id: _scale.id
                        }
                    }
                    
                    return 1;
                }
            })
        }
    },
    getScaleByColumnName: function(column, encode){
        let columnName = column.columnname;
        if(this.columnInfo[columnName]){
            let _scale_id = this.columnInfo[columnName]["scale_id"];
            if(defined(_scale_id)){
                if(isArray(_scale_id)){
                    let scale;
                    _scale_id.map(d => {
                        let scaleConf = this.scales[d].conf;
                        if(scaleConf.encode && scaleConf.encode == encode){
                            scale =  this.scales[d];
                        }
                    })
                    return scale || this.scales[_scale_id[0]];
                }else if(this.scales[_scale_id]){
                    return this.scales[_scale_id];
                }else{
                    return null;
                } 
            }
            return null;
        }
    },
    getColumnByColumnName: function(columnName){
        let column;
        this.spec.columns.filter(d => {
            if(d.columnname == columnName)
                column = d;
                return 1;
        });
        return column;
    }
};

Registry.setComponent('viewParser', ViewFlow);
