// @zohocharts/zc/src/parser/parser-utils.js
//
import { mergeJSON, invokeFunction, getPropVal, isArray } from '../utils/zcutils';
import { isFunction } from '../utils/function-utils';
import { isExpressionType, getExpressionValue } from '../mark/parser/parser-helpers';

export function extendScaleParser(defaultParser, customParser, chartObj) {
    if (isFunction(customParser) || (typeof customParser === 'string' && isFunction(getPropVal(window, customParser)))) {
        return invokeFunction(customParser, defaultParser);
    }

    if (customParser.extend) {
        let expressionValues = applyExpressionValueForParser(customParser, chartObj.viewFlow);
        defaultParser = mergeJSON(defaultParser, customParser, true, true, true);

        expressionValues.map(d => {
            defaultParser[d[0]] = d[1];
        })

    } else {
        if (customParser.domain) {
            if (customParser.domain.extend) {
                defaultParser.domain = mergeJSON(defaultParser.domain, customParser.domain, true, true, true);
            } else {
                defaultParser.domain = customParser.domain;
            }
        }
        if (customParser.range) {
            if (customParser.range.extend) {
                defaultParser.range = mergeJSON(defaultParser.range, customParser.range, true, true, true);
            } else {
                defaultParser.range = customParser.range;
            }
        }
    }
    return defaultParser;
}


export function applyExpressionValueForParser(parser, viewFlow){
    let attr_values = new Array(0);
    //For a custom parser with expression, apply expression and sen the values alone
    for (let attr in parser){
        if(isExpressionType(parser[attr])){
            let config = parser[attr],
                value = getExpressionValue(null, [1,1], config, null, viewFlow);
                attr_values.push([attr, value]);

            if(config.args && config.args.reacts){
                //As some of the property in scale has react varuiable, add scale component as dependency to trigger react 
                addScaleAsDependencyToReactTrigger(config.args, parser, viewFlow);
            }
        }
    }

    return attr_values;
}

function addScaleAsDependencyToReactTrigger(config, parser, viewFlow){
    let reactsId = config.reacts,
        reactsIds = isArray(reactsId) ? reactsId : [reactsId];

        reactsIds.map(id => {
            let mainConfig = viewFlow.reacts[id];
            if(parser && !mainConfig.dependencyIds.has(parser.id)){
                mainConfig.dependencies.push({
                    recalculate: true,
                    component: "scale",
                    id: parser.id
                })
                mainConfig.dependencyIds.add(parser.id);
            }
            
        })
}
