// @zohocharts/zc/src/parser/parser-reactive-flow.js
import { d3_set } from "../core/d3-constants";
import { isExpressionType } from "../mark/parser/parser-helpers";
import { isArray } from "../utils/zcutils";

export var CONFIG_WITH_LAYOUT_RECALCULATION = d3_set();

CONFIG_WITH_LAYOUT_RECALCULATION.add(["x", "y", "width", "height", "size", "scale"]);

export var CONFIG_WITHOUT_LAYOUT_RECALCULATION = d3_set();

CONFIG_WITHOUT_LAYOUT_RECALCULATION.add(["fill", "fillColor", "fillOpacity", "stroke", "strokeColor", "strokeOpacity"]);//Still need to add fully

export function collectReactiveDependencies(prop, value, config, obj, parent){
    if(isExpressionType(config[prop])){
        let viewFlow = obj.__tokens__ || obj.cache.__tokens__,
            scalesId = config[prop].args && config[prop].args.scales,
            reactsId = config[prop].args && config[prop].args.reacts;
        if(scalesId){
            collectScaleDependencies(prop, value, scalesId, viewFlow, obj, parent);
        }

        //Add properties dependent on react variable
        if(reactsId){
            collectReactDependecies(prop, value, reactsId, viewFlow, obj, parent);
        }
    }
}

function collectReactDependecies(prop, value, reactsId, viewFlow, obj, parent){
    let  reactsIds = isArray(reactsId) ? reactsId : [reactsId];

        reactsIds.map(id => {
            let mainConfig = viewFlow.reacts[id];
            if(obj.parser){
                let parserId = generateParserId(obj.parser);
                if(!mainConfig.dependencyIds.has(parserId)){
                    let config = {}
                    config[prop] = value;
                    mainConfig.dependencies.push({
                        parser: obj.parser,
                        id: parserId,
                        parent,
                        config,
                        recalculate: CONFIG_WITH_LAYOUT_RECALCULATION.has(prop)
                    })
                    mainConfig.dependencyIds.add(obj.parser.id);
                }else{
                    let existingDependency = mainConfig.dependencies.filter(m => m.id == parserId);
                    if(existingDependency[0] && !existingDependency[0].recalculate){
                        existingDependency[0].config[prop] = value;
                    }
                }
                
            }
        })
}

function collectScaleDependencies(prop, value, scalesId, viewFlow, obj, parent){
    let scalesIds = isArray(scalesId) ? scalesId : [scalesId];
        //Add properties dependent on scale
        scalesIds.map(id => {
            if(obj.parser && !viewFlow.scales[id].dependencyIds.has(obj.parser.id)){
                viewFlow.scales[id].dependencies.push({
                    type: "mark",
                    parent,
                    parser: obj.parser,
                    prop
                })
                viewFlow.scales[id].dependencyIds.add(obj.parser.id);
            }
            
        })
}

function generateParserId(parser){
    if(parser.id){
        return parser.id
    }
    else if(parser.mark && parser.selector){
        return parser.mark+parser.selector;
    }
    
}
export function passNotificationTodependencies(){
    
}
