// @zohocharts/zc/src/threshold/thresholdRange.js
//$Id$
//thresholdRange.js
import { axis } from '../core/main-zc-init';
import { isObjectPropDefined, defined, isArray, validateNullVal, decodeHTML, utils, getPropVal, bbox } from '../utils/zcutils';
import { overrideThresholdFonts, get_threshold_text_orient } from './threshold-utils';
import { appendEle } from '../utils/zcsvgutils';
import { TIME_DATATYPE, ORDINAL_DATATYPE } from '../core/datatype-constants';
import { generateDateVal } from '../formatter/formatter';
import { d3_select } from '../core/d3-constants';
import { getCustomThresholdText } from './customThresholdText';
import { trimText } from '../utils/zcsvgutils';
import { getDashArray, getLineCap } from '../utils/dashstyles';
import { mathMax, mathMin } from '../core/zc-constants';
import { $ZCG } from '../core/zcglobal';
import {getAxisObject} from "../axis/axisutils";

export * from './threshold-init';

axis.threshold.range = function() {
    var index, chartObj, type, parentObj;

    function chart(selection) {
        selection.each(function(data) {
            var container = d3_select(this);

            chartObj = parentObj.chartObj();
            index = parentObj.thresholdIndex();
            type = parentObj.type();

            var systemconf = chartObj.systemConf;
            var label = data.label;
            label = overrideThresholdFonts(
                label,
                isObjectPropDefined(
                    chartObj.userdata,
                    `chart.axes.${type}axis${type === 'x' ? '' : `.${index}`}.threshold.range.label`,
                    true
                )
            );
            var fillColor = [],
                fillOpacity = [],
                strokeColor = [],
                strokeDashArray = [],
                lineCap = [],
                hAlign = [],
                thName = [],
                fontColor = [],
                fontSize = [],
                fontFamily = [],
                fontWeight = [],
                fontStyle = [];
            var rect = chartObj.container.node().getBoundingClientRect(),
                plotarea = chartObj.plotarea,
                bound = {
                    left: rect.left + plotarea.left,
                    top: rect.top + plotarea.top,
                    width: plotarea.width,
                    height: plotarea.height,
                    right: rect.left + plotarea.left + plotarea.width,
                    bottom: rect.top + plotarea.top + plotarea.height
                },
                orient = chartObj.axes[type].orient,
                actualOrient = parentObj.axisObj().component.orient(),
                margin = {
                    top: 0,
                    left: 0,
                    right: 0,
                    bottom: 0
                },
                overFlow = '..';

            var wrap = appendEle(container, 'g', data.show ? [data] : [], '.', 'class', `thresholdrangegroup-${type}`);

            var axesRotated = chartObj.axes.rotated;

            if (data.show) {
                var datatype = chartObj.dataset.getDataType(type, index);
                var formatConf;
                var colInfo = chartObj.dataset.getColumn(type, index);
                if (datatype == TIME_DATATYPE) {
                    formatConf = colInfo.time != null ? colInfo.time : null;
                }
                var values = [],
                    arr = [];

                var labelPadding = {
                    x: 2,
                    y: 2
                };
                var xAxisLine = systemconf.chart.axes.xaxis.axisline,
                    yAxisLine = systemconf.chart.axes.yaxis[index].axisline;

                if (defined(xAxisLine) && xAxisLine.show) {
                    labelPadding.x += +xAxisLine.strokeWidth - 1;
                }

                if (defined(yAxisLine) && yAxisLine.show) {
                    labelPadding.y += +yAxisLine.strokeWidth - 1;
                }
                let axisObj_x = getAxisObject(chartObj, 'x', index),
                    axisObj_y = getAxisObject(chartObj, 'y', index);
                var scale_x = !axesRotated ? axisObj_x.scale : axisObj_y.scale;
                var scale_y = !axesRotated ? axisObj_y.scale : axisObj_x.scale;

                var isBarEnabled = chartObj.dataObject.noofBarEnabledSeries;
                var rangeBand_x = scale_x.bandwidth != null && isBarEnabled ? scale_x.bandwidth() : 0,
                    rangeBand_y = scale_y.bandwidth != null && isBarEnabled ? scale_y.bandwidth() : 0,
                    strangeBand_x = scale_x.bandwidth != null && !isBarEnabled ? scale_x.bandwidth() / 2 : 0,
                    strangeBand_y = scale_y.bandwidth != null && !isBarEnabled ? scale_y.bandwidth() / 2 : 0;

                var count = 0;
                var domain = ((type == 'x' && !axesRotated) || (type == 'y' && axesRotated) ? scale_x : scale_y).domain();
                var domainLength = domain.length - 1;

                data.values.map(function(d, i) {
                    arr =
                        d != null && d != ''
                            ? d.map(function(dt) {
                                  return datatype == TIME_DATATYPE ? generateDateVal(dt, null, formatConf).getTime() : dt;
                              })
                            : null;
                    var rangeCheck =
                        datatype == ORDINAL_DATATYPE
                            ? domain.indexOf(arr[0]) !== -1 && domain.indexOf(arr[1]) !== -1
                            : !(arr[0] > domain[domainLength] || arr[1] < domain[0]);
                    if (arr && rangeCheck) {
                        /*
                         * Range should be within plot domain
                         */
                        if (datatype !== ORDINAL_DATATYPE) {
                            arr[0] = mathMax(arr[0], domain[0]);
                            arr[1] = mathMin(arr[1], domain[domainLength]);
                        }
                        values[count] = arr;

                        // rangeBoxRect properties
                        fillColor[count] = (isArray(data.fillColor) ? data.fillColor[i] : data.fillColor) || 'grey';
                        fillOpacity[count] = validateNullVal(isArray(data.fillOpacity) ? data.fillOpacity[i] : data.fillOpacity, '0.2');

                        strokeColor[count] = (isArray(data.strokeColor) ? data.strokeColor[i] : data.strokeColor) || 'black';
                        strokeDashArray[count] = isArray(data.dashStyle) ? data.dashStyle[i] : data.dashStyle;
                        lineCap[count] = isArray(data.lineCap) ? data.lineCap[i] : data.lineCap;

                        // rangeBoxText properties
                        hAlign[count] = (isArray(data.label.hAlign) ? data.label.hAlign[i] : data.label.hAlign) || 'center';
                        thName[count] = getCustomThresholdText(getPropVal(data, `label.text[${i}]`) || '', data.values, colInfo, chartObj);

                        fontColor[count] = isArray(label.fontColor) ? label.fontColor[i] : label.fontColor;
                        fontStyle[count] = isArray(label.fontStyle) ? label.fontStyle[i] : label.fontStyle;
                        fontFamily[count] = isArray(label.fontFamily) ? label.fontFamily[i] : label.fontFamily;
                        fontWeight[count] = isArray(label.fontWeight) ? label.fontWeight[i] : label.fontWeight;
                        fontSize[count] =
                            label.fontSize instanceof Function || isArray(label.fontSize) ? label.fontSize[i] : label.fontSize;

                        count++;
                    }
                });

                var rangeBoxgroup = appendEle(wrap, 'g', values, '.', 'class', `rangebox-${index}`);

                rangeBoxgroup.append('rect');
                rangeBoxgroup.append('text');

                var rangeBoxRect = rangeBoxgroup.select('rect');
                var rangeBoxText = rangeBoxgroup.select('text');

                rangeBoxRect
                    .attrs({
                        x: function(d, i) {
                            return (type == 'x' && !axesRotated) || (type == 'y' && axesRotated)
                                ? scale_x(d[0]) + strangeBand_x + chartObj.plotarea.left
                                : chartObj.plotarea.left + labelPadding[type];
                        },
                        y: function(d, i) {
                            return (type == 'x' && !axesRotated) || (type == 'y' && axesRotated)
                                ? chartObj.plotarea.top
                                : scale_y(d[1]) + strangeBand_y + chartObj.plotarea.top;
                        },
                        width: function(d, i) {
                            return (type == 'x' && !axesRotated) || (type == 'y' && axesRotated)
                                ? scale_x(d[1]) - scale_x(d[0]) + rangeBand_x
                                : chartObj.plotarea.width - labelPadding[type];
                        },
                        height: function(d, i) {
                            return (type == 'x' && !axesRotated) || (type == 'y' && axesRotated)
                                ? chartObj.plotarea.height - labelPadding[type]
                                : scale_y(d[0]) - scale_y(d[1]) + rangeBand_y;
                        }
                    })
                    .styles({
                        fill: function(d, i) {
                            return fillColor[i];
                        },
                        'fill-opacity': function(d, i) {
                            return fillOpacity[i];
                        },
                        stroke: function(d, i) {
                            return strokeColor[i];
                        },
                        'stroke-dasharray': function(d, i) {
                            return getDashArray(strokeDashArray[i]);
                        },
                        'stroke-linecap': function(d, i) {
                            return getLineCap(lineCap[i], strokeDashArray[i]);
                        }
                    });

                rangeBoxText
                    .attrs({
                        transform: function(d, i) {
                            var box = bbox(rangeBoxRect.nodes()[i]);
                            var transY = axis.threshold.textAlign(
                                hAlign[i],
                                type,
                                chartObj.plotarea,
                                axesRotated,
                                labelPadding[type],
                                orient
                            );
                            if ((type == 'x' && !axesRotated) || (type == 'y' && axesRotated)) {
                                return `translate(${box.x + box.width / 2},${transY + chartObj.plotarea.top}) rotate(-90)`;
                            }
                            return `translate(${transY + chartObj.plotarea.left},${box.y + box.height / 2})`;
                        },
                        dy: '0.3em'
                    })
                    .styles({
                        'font-family': function(d, i) {
                            return fontFamily[i] || systemconf.chart.fontFamily || systemconf.canvas.fontFamily || $ZCG.fontFamily;
                        },
                        'font-size': function(d, i) {
                            return `${utils.getVal(fontSize[i] || systemconf.chart.fontSize || systemconf.canvas.fontSize)}px`;
                        },
                        fill: function(d, i) {
                            return fontColor[i] || systemconf.chart.fontColor || systemconf.canvas.fontColor || $ZCG.fontColor;
                        },
                        'font-weight': function(d, i) {
                            return fontWeight[i] || systemconf.chart.fontWeight || systemconf.canvas.fontWeight || $ZCG.fontWeight;
                        },
                        'font-style': function(d, i) {
                            return fontStyle[i] || systemconf.chart.fontStyle || systemconf.canvas.fontStyle || $ZCG.fontStyle;
                        },
                        'text-Anchor': function(d, i) {
                            return hAlign[i] == 'center' ? 'middle' : hAlign[i] == 'right' ? 'end' : 'start';
                        }
                    })
                    .text(function(d, i) {
                        return decodeHTML(thName[i]);
                    })
                    .each(function() {
                        trimText(
                            this,
                            this.textContent,
                            this.getBoundingClientRect(),
                            bound,
                            get_threshold_text_orient(actualOrient),
                            overFlow,
                            margin
                        );
                    });
            }
        });
    }

    chart.parentObj = function(_) {
        if (!arguments.length) {
            return parentObj;
        }
        parentObj = _;
        return chart;
    };
    return chart;
};

export var _thresholdRange = axis.threshold.range;
