// @zohocharts/zc/src/threshold/thresholdLine.js
//$Id$
//thresholdLine.js
import { axis, events } from '../core/main-zc-init';
import { isObjectPropDefined, getPropVal, isArray, defined, pick, mergeJSON, utils, decodeHTML, bbox } from '../utils/zcutils';
import { isFunction } from '../utils/function-utils';
import { generateDateVal } from '../formatter/formatter';
import { d3_select } from '../core/d3-constants';
import { overrideThresholdFonts } from './threshold-utils';
import { TIME_DATATYPE } from '../core/datatype-constants';
import { getTriggeringOffset } from '../events/events-utils';
import { trimText, appendEle } from '../utils/zcsvgutils';
import { getDashArray } from '../utils/dashstyles';
import { get_threshold_text_orient } from './threshold-utils';
import { TRANSPARENT, NULL } from '../core/zc-constants';
import { getOppositeOrient } from '../axis/axisutils';
import { getFormattedValue } from '../formatter/formatter';
import { getSVGsymbol } from '../utils/svg-symbols';
import { mathMin } from '../core/zc-constants';
import { getCustomThresholdText } from './customThresholdText';
import { $ZCG } from '../core/zcglobal';
import { HIGHLIGHT_CONTEXT } from '../events/visualEffects/highlight/highlight-context';

export * from './threshold-init';

axis.threshold.line = function() {
    var index, plotarea, systemconf, chartObj, axesRotated, type, parentObj, axisObj, values;

    function chart(selection) {
        selection.each(function(data) {
            var container = d3_select(this);
            chart.update = function() {
                container.transition().call(chart);
            };

            chartObj = parentObj.chartObj();
            index = parentObj.thresholdIndex();
            type = parentObj.type();
            axisObj = parentObj.axisObj();
            values = parentObj.values();

            plotarea = chartObj.plotarea;
            var scaleObj = axisObj.scale;
            var range = scaleObj.range();
            range = range.slice(0, range.length);
            range.sort(function(a, b) {
                return a - b;
            });
            var bandwidth = scaleObj.bandwidth != null ? scaleObj.bandwidth() / 2 : 0;
            var label = data.label;
            label = overrideThresholdFonts(
                label,
                isObjectPropDefined(
                    chartObj.userdata,
                    `chart.axes.${type}axis${type === 'x' ? '' : `.${index}`}.threshold.line.label`,
                    true
                )
            );
            var markerInfo = data.marker || {};
            var markerEnabled = pick(markerInfo.enabled, true);
            axesRotated = chartObj.axes.rotated;
            systemconf = chartObj.systemConf;
            var thval = [],
                thname = [],
                thclr = [];

            // var yAxisColumnOrder = chartObj.systemConf.metadata.axes[type];
            // var colInfo = chartObj.systemConf.metadata.columns[index];

            var labelPadding = {
                x: 2,
                y: 2
            };
            var xAxisLine = systemconf.chart.axes.xaxis.axisline,
                yAxisLine = systemconf.chart.axes.yaxis[index].axisline,
                effect = chartObj.filterEffectsID;

            if (defined(xAxisLine) && xAxisLine.show) {
                labelPadding.x += +xAxisLine.strokeWidth - 1;
            }

            if (defined(yAxisLine) && yAxisLine.show) {
                labelPadding.y += +yAxisLine.strokeWidth - 1;
            }

            var count = 0;

            var hAlign = [],
                strokeDashArray = [],
                strokeWidth = [],
                markerWidth = [],
                markerBBox = [],
                markerSize = [],
                fontSizes = [],
                fontFamilies = [];

            var datatype = chartObj.dataset.getDataType(type, index);
            var formatConf,
                eventDisabled = getPropVal(data.events, 'enabled') !== false;

            var colInfo = chartObj.dataset.getColumn(type, index),
                columnindex = chartObj.dataset.getColumnIdx(type, index);
            if (datatype == TIME_DATATYPE) {
                formatConf = getPropVal(colInfo, TIME_DATATYPE) || null;
            }

            let hierarchical = chartObj.dataObject.hierarchical,
                hasMultiCluster,
                rowCount,
                columnCount,
                grid;
            if (hierarchical) {
                let rowList = hierarchical.pivotLayout.getRootNodeList('row'),
                    columnList = hierarchical.pivotLayout.getRootNodeList('column');
                (rowCount = rowList.getCount()), (columnCount = columnList.getCount()), (hasMultiCluster = rowList.hasMultiCluster()), grid;
            }

            if (hasMultiCluster && !rowCount && !columnCount) {
                grid = `${index}_${0}`;
            }

            values.map(function(d, i) {
                var d_scale =
                    defined(d) && d !== '' ? (datatype == TIME_DATATYPE ? generateDateVal(d, null, formatConf).getTime() : d) : null;    // #ZC4745
                if (
                    d != null &&
                    d !== '' &&
                    scaleObj(d_scale) + bandwidth >= range[0] &&
                    scaleObj(d_scale) + bandwidth <= range[range.length - 1]
                ) {
                    thval[count] = d_scale;
                    thname[count] = getCustomThresholdText(getPropVal(data, `label.text[${i}]`) || '', values, colInfo, chartObj);
                    thclr[count] = (isArray(data.color) ? data.color[i] : data.color) || 'black';
                    hAlign[count] = defined(data.label)
                        ? (isArray(data.label.hAlign) ? data.label.hAlign[i] : data.label.hAlign) || 'center'
                        : 'center';
                    strokeWidth[count] = (isArray(data.strokeWidth) ? data.strokeWidth[i] : data.strokeWidth) || 1;
                    strokeDashArray[count] = isArray(data.dashStyle) ? data.dashStyle[i] : data.dashStyle;
                    markerWidth[count] =
                        defined(data.marker) && data.marker.enabled
                            ? (isArray(data.marker.strokeWidth) ? data.marker.strokeWidth[i] : data.marker.strokeWidth) || 2
                            : null;
                    markerSize[count] =
                        data.marker && data.marker.enabled
                            ? (isArray(data.marker.size) ? data.marker.size[i] : data.marker.size) || 1
                            : null;
                    count++;
                }
            });
            var rect = chartObj.container.node().getBoundingClientRect(),
                bound = {
                    left: rect.left + plotarea.left,
                    top: rect.top + plotarea.top,
                    width: plotarea.width,
                    height: plotarea.height,
                    right: rect.left + plotarea.left + plotarea.width,
                    bottom: rect.top + plotarea.top + plotarea.height
                },
                boundLeft = bound.left,
                boundTop = bound.top,
                orient = axisObj.component.orient(),
                margin = {
                    top: 0,
                    left: 0,
                    right: 0,
                    bottom: 0
                },
                overFlow = '..',
                key = `threshold${type}${index}`,
                thresholdPoints = (chartObj.cache.thresholdCustomPoints = pick(chartObj.cache.thresholdCustomPoints, {})),
                thresholdEvents = (thresholdPoints[key] = pick(thresholdPoints[key], [])),
                thresholdTriggerOffset = pick(getThresholdTriggerOffset(data, chartObj), 10);
            thresholdEvents.superKey = 'threshold';
            thresholdEvents.triggerOffset = thresholdTriggerOffset;
            thresholdEvents.grid = grid;
            container.attr('transform', `translate(${plotarea.left},${plotarea.top})`);
            var wrap = appendEle(container, 'g', [data], '.', 'class', `thresline-${index}`);
            //plotarea border size to include in markerbox width
            var plotareaBorderEle = chartObj.staticPaneSVG.selectAll('g.plotareaBorder-property').node();
            var plotareaBorder_bkp = plotareaBorderEle.border || plotareaBorderEle[`border-${orient}`];
            // var plotareaBorder_Oppbkp = (orient == "left") ? (plotareaBorderEle["border"] || plotareaBorderEle["border-right"]) : (orient == "right") ? (plotareaBorderEle["border"] || plotareaBorderEle["border-left"]) : null; //For dual side axes, need to consider border size on both axes
            var plotareaBorder_Oppbkp = plotareaBorderEle.border || plotareaBorderEle[`border-${getOppositeOrient(orient)}`];
            var wid1 = plotareaBorder_bkp && plotareaBorder_bkp.clr !== TRANSPARENT ? plotareaBorder_bkp.size : 0;
            var wid2 = plotareaBorder_Oppbkp && plotareaBorder_Oppbkp.clr !== TRANSPARENT ? plotareaBorder_Oppbkp.size : 0;
            var plotareaBorderWid_bkp = wid1 + wid2;

            var width = pick(chartObj.defaultScroll && chartObj.defaultScroll.dimension && chartObj.defaultScroll.dimension.width, plotarea.width),
                height = pick(chartObj.defaultScroll && chartObj.defaultScroll.dimension && chartObj.defaultScroll.dimension.height, plotarea.height);
            var marker = appendEle(wrap, 'path', markerInfo != null && markerEnabled ? thval : [], '.', 'class', 'marker');
            marker
                .attr('d', function(d, i) {
                    return getSVGsymbol(isArray(markerInfo.symbol) ? markerInfo.symbol[i] : markerInfo.symbol)(markerSize[i] * 10);
                })
                // .attr("transform",function(d,i){return "translate("+((type == "x" && !axesRotated)||(type == "y" && axesRotated)? (scale(d)+bandwidth) : (0+(((markerWidth[i] || 1)*10)/2))) + "," + ((type == "x" && !axesRotated)||(type == "y" && axesRotated)? (plotarea.height-(((markerWidth[i] || 1)*10)/2)) :scale(d)+bandwidth)+")"})
                .style('stroke-width', function(d, i) {
                    return markerWidth[i] || 2;
                })
                .style('stroke', function(d, i) {
                    return (isArray(markerInfo.strokeColor) ? markerInfo.strokeColor[i] : markerInfo.strokeColor) || thclr[i];
                })
                .attr('fill', function(d, i) {
                    var fillClr = isArray(markerInfo.fillColor) ? markerInfo.fillColor[i] : markerInfo.fillColor;
                    return fillClr || (fillClr === null ? thclr[i] : 'white'); //If 'fillColor' is given as null explicitly by the user, use thclr
                }); 
            var markerOrient = markerInfo.orient;
            marker.attr('transform', function(d, i) {
                var mrkrWidth = markerWidth[i] / 2;
                var bbox_marker = bbox(this);
                markerBBox[i] = bbox_marker.width + wid1;
                
                var translateX =
                        orient == 'bottom' || orient == 'top'
                            ? scaleObj(d) + bandwidth
                            : orient == 'left'
                            ? mrkrWidth + wid1 + bbox_marker.width / 2
                            : width - (mrkrWidth + wid1 + bbox_marker.width / 2),
                    // translateY = (orient == "bottom") ? (plotarea.height - mrkrWidth - ((plotareaBorderWid_bkp + ((bbox.width) / 2)))) : (orient == "top") ? (mrkrWidth + plotareaBorderWid_bkp + ((bbox.height) / 2)) : (scale(d) + bandwidth);
                    translateY =
                        orient == 'bottom'
                            ? height - bbox_marker.width / 2 - wid1 - mrkrWidth
                            : orient == 'top'
                            ? mrkrWidth + bbox_marker.height / 2 + wid1
                            : scaleObj(d) + bandwidth;

                if(markerOrient){
                    if (orient == 'bottom' || orient == 'top') {
                        translateX = scaleObj(d) + bandwidth;
                        translateY =
                            markerOrient == 'bottom'
                                ? height - bbox_marker.width / 2 - wid1 - mrkrWidth
                                : markerOrient == 'top'
                                ? mrkrWidth + bbox_marker.height / 2 + wid1
                                : translateY;
                    }
                    if (orient == 'left' || orient == 'right') {
                        translateY = scaleObj(d) + bandwidth;
                        translateX =
                            markerOrient == 'left'
                                ? mrkrWidth + wid1 + bbox_marker.width / 2
                                : markerOrient == 'right'
                                ? width - (mrkrWidth + wid1 + bbox_marker.width / 2)
                                : translateX;
                    }
                }
                if (eventDisabled && events.customPoints) {
                    var thresholdData = function(chartObj, index) {
                        return function() {
                            var cache = chartObj.cache.thresholdCustomPoints[key][index];

                            return {
                                _x: cache && cache._x,
                                _y: cache && cache._y,
                                _size: cache && cache._size,
                                _visible: true
                            };
                        };
                    };
                    thresholdEvents.push({
                        _data: thresholdData(chartObj, i), //yaxisIndex+itemindex
                        _x: translateX,
                        _y: translateY,
                        _key: `threshold_${chartObj.id}_${type}_${index}_${i}`,
                        width: markerBBox[i],
                        height: markerBBox[i],
                        color: thclr[i],
                        text: thname[i],
                        type: 'threshold-marker',
                        triggerOffset: thresholdTriggerOffset,
                        _size: (markerSize[i] || markerWidth[i]) * 10 + markerWidth[i] / 2,
                        _mSize: markerSize[i],
                        marker: markerInfo,
                        point: values[i],
                        formattedValue: getFormattedValue(
                            colInfo[colInfo.datatype],
                            values[i],
                            colInfo.datatype,
                            chartObj,
                            'threshold',
                            false
                        ),
                        // value : d,
                        // columnindex : columnindex,
                        thresholdType: type,
                        columnIndex: columnindex,
                        columnInfo: colInfo,
                        yColumnIndex: index,
                        itemIndex: i,
                        events: data.events || {},
                        disabled: function() {
                            return container.__data__ != NULL || !chartObj.cache.thresholdCustomPoints[key];
                        },
                        _defaultH: function() {
                            return [
                                {
                                    shape: 'marker',
                                    path: function() {
                                        return chartObj.eventHandler.CustomPoints.hoverPoint(this);
                                    },
                                    context: function() {
                                        return HIGHLIGHT_CONTEXT.get('scatter').call(this, chartObj);
                                    }
                                }
                            ];
                        }
                    });
                }
                return `translate(${translateX},${translateY})`;
            });

            var line = appendEle(wrap, 'line', thval);
            line.attr('x1', function(d, i) {
                if((markerOrient == 'left' || markerOrient == 'right') && (orient == 'left' || orient == 'right')){
                    return markerOrient == 'left' ? (markerBBox[i] || 0) + markerWidth[i] : wid2;
                }
                return orient == 'bottom' || orient == 'top'
                    ? scaleObj(d) + bandwidth
                    : orient == 'left'
                    ? (markerBBox[i] || 0) + markerWidth[i]
                    : wid2;
            })
                .attr('y1', function(d, i) {
                    if((markerOrient == 'bottom' || markerOrient == 'top') && (orient == 'bottom' || orient == 'top')){
                        return markerOrient == 'bottom' ? wid1 : (markerBBox[i] || 0) + markerWidth[i];
                    }

                    return orient == 'bottom' ? wid1 : orient == 'top' ? (markerBBox[i] || 0) + markerWidth[i] : scaleObj(d) + bandwidth;
                })
                .attr('x2', function(d, i) {
                    if ((markerOrient == 'left' || markerOrient == 'right') && (orient == 'left' || orient == 'right')) {
                        return markerOrient == 'left' ? width - wid2 : width - markerWidth[i] - (markerBBox[i] || 0);
                    }
                    let x2 =
                        orient == 'bottom' || orient == 'top'
                            ? scaleObj(d) + bandwidth
                            : orient == 'left'
                            ? width - wid2
                            : width - markerWidth[i] - (markerBBox[i] || 0);

                    return effect ? x2 + 1 : x2;
                })
                .attr('y2', function(d, i) {
                    if((markerOrient == 'bottom' || markerOrient == 'top') && (orient == 'bottom' || orient == 'top')){
                        return markerOrient == 'bottom' ?  height - markerWidth[i] - (markerBBox[i] || 0) : height - wid2;
                    }
                    let y2 =
                        orient == 'bottom'
                            ? height - markerWidth[i] - (markerBBox[i] || 0)
                            : orient == 'top'
                            ? height - wid2
                            : scaleObj(d) + bandwidth;
                    return effect ? y2 + 1 : y2;
                })
                .attr('stroke', function(d, i) {
                    return `    ${thclr[i]}`;
                })
                .attr('stroke-width', function(d, i) {
                    return `${strokeWidth[i]}px`;
                })
                .attr('shaperendering', 'crispedges')
                .attr('stroke-dasharray', function(d, i) {
                    return getDashArray(strokeDashArray[i], strokeWidth[i]);
                });

            var text = appendEle(wrap, 'text', label != null ? thval : []);
            text.text(function(d, i) {
                return decodeHTML(thname[i]);
            })
                .attr('x', function(d, i) {
                    var textPadding = markerSize[i] * 10 + labelPadding[type] + (markerWidth[i] || 2);
                    var x =
                        (type == 'x' && !axesRotated) || (type == 'y' && axesRotated)
                            ? -axis.threshold.textAlign(hAlign[i] || 'right', type, plotarea, axesRotated, textPadding, orient)
                            : axis.threshold.textAlign(hAlign[i] || 'right', type, plotarea, axesRotated, textPadding, orient);
                    return x;
                })
                .attr('y', function(d, i) {
                    var y =
                        (type == 'x' && !axesRotated) || (type == 'y' && axesRotated) ? scaleObj(d) + bandwidth : scaleObj(d) + bandwidth;
                    return y - 5;
                })

                .attr('transform', (type == 'x' && !axesRotated) || (type == 'y' && axesRotated) ? 'rotate(-90)' : 'rotate(0)')
                .styles({
                    'font-family': function(d, i) {
                        fontFamilies[i] =
                            (isArray(label.fontFamily) ? label.fontFamily[i] : label.fontFamily) ||
                            systemconf.chart.fontFamily ||
                            systemconf.canvas.fontFamily ||
                            $ZCG.fontFamily;
                        return fontFamilies[i];
                    },
                    'font-size': function(d, i) {
                        fontSizes[i] = utils.getVal(
                            (isFunction(label.fontSize) || isArray(label.fontSize) ? label.fontSize[i] : label.fontSize) ||
                                systemconf.chart.fontSize ||
                                systemconf.canvas.fontSize
                        );
                        return `${fontSizes[i]}px`;
                    },
                    fill: function(d, i) {
                        return (
                            (isArray(label.fontColor) ? label.fontColor[i] : label.fontColor) ||
                            systemconf.chart.fontColor ||
                            systemconf.canvas.fontColor ||
                            $ZCG.fontColor
                        );
                    },
                    'font-weight': function(d, i) {
                        return (
                            (isArray(label.fontWeight) ? label.fontWeight[i] : label.fontWeight) ||
                            systemconf.chart.fontWeight ||
                            systemconf.canvas.fontWeight ||
                            $ZCG.fontWeight
                        );
                    },
                    'font-style': function(d, i) {
                        return (
                            (isArray(label.fontStyle) ? label.fontStyle[i] : label.fontStyle) ||
                            systemconf.chart.fontStyle ||
                            systemconf.canvas.fontStyle ||
                            $ZCG.fontStyle
                        );
                    },
                    'text-Anchor': function(d, i) {
                        if (orient == 'left' || orient == 'bottom') {
                            return hAlign[i] == 'center' ? 'middle' : hAlign[i] == 'right' ? 'end' : 'start';
                        } else {
                            return hAlign[i] == 'center' ? 'middle' : hAlign[i] == 'right' ? 'start' : 'end';
                        }
                    }
                })
                .each(function(d, i) {
                    var text = decodeHTML(thname[i]);
                    var actualMarkerSize = markerSize[i] * 10,
                        textWidth = utils.fontUtils.getTextWidth(text, fontSizes[i], fontFamilies[i]) + actualMarkerSize,
                        size = orient === 'top' || orient === 'bottom' ? plotarea.height : plotarea.width,
                        adjustment;

                    if (textWidth > size) {
                        adjustment = mathMin(textWidth - size, actualMarkerSize) + plotareaBorderWid_bkp;
                        bound.left = boundLeft + adjustment;
                        bound.top = boundTop + adjustment;
                        trimText(
                            this,
                            this.textContent,
                            this.getBoundingClientRect(),
                            bound,
                            get_threshold_text_orient(orient),
                            overFlow,
                            margin
                        );
                    }
                });
        });
    }

    chart.parentObj = function(_) {
        if (!arguments.length) {
            return parentObj;
        }
        parentObj = _;
        return chart;
    };

    chart.getMetaInfo = function(point) {
        return {
            itemIndex: point.itemIndex,
            point: point.point,
            data: mergeJSON(chartObj.systemConf.metadata.columns[point.columnIndex], {
                columnindex: point.columnIndex
            })
        };
    };

    /**
     *
     * @param {Object} markerInfo
     * @param {Chart} chartInstance
     */
    function getThresholdTriggerOffset(markerInfo, chartInstance) {
        return pick(getPropVal(markerInfo, 'events.triggerOffset'), getTriggeringOffset(chartInstance, chartInstance.chartTypes[0].value));
    }

    // function align(hAlign,type,data){
    // 	var x;
    // 	if (hAlign == "left"){
    // 		if(axesRotated){
    // 			x = (type == "x")?0:plotarea.height;
    // 		}else{
    // 			x = (type == "x")?plotarea.height:0;
    // 		}
    // 	}else if(hAlign == "center"){
    // 		if(axesRotated){
    // 			x = (type == "x")?plotarea.width/2:plotarea.height/2;
    // 		}else{
    // 			x = (type == "x")?plotarea.height/2:plotarea.width/2;
    // 		}
    // 	}else if(hAlign == "right"){
    // 		if(axesRotated){
    // 			x= (type == "x")?plotarea.width:0;
    // 		}else{
    // 			x= (type == "x")?0:plotarea.width;
    // 		}
    // 	}
    // 	return x
    // }
    return chart;
};

export var _thresholdLine = axis.threshold.line;
