// @zohocharts/zc/src/tooltip/tooltip.js
//$Id$
//tooltip.js
import { Registry } from '../core/zc-registry';
import { events } from '../core/main-zc-init';
import {
    isObjectPropDefined,
    mergeJSON,
    validateNullVal,
    defined,
    shorthand_px,
    invokeFunction,
    encodeHTML,
    isArray,
    isUndefined,
    pInt,
    getBorderRadius,
    isHTMLElement,
    perToPx,
    getDepth
} from '../utils/zcutils';
import { getFontSizeByHierarchy } from '../utils/zcdatautils';
import { appendEle, applyFont } from '../utils/zcsvgutils';
import { NONE } from '../core/zc-constants';
import { $ZC_TOOLTIP_POSITION } from './tooltip-position';
import { getRawColor } from '../utils/colorutils';
import { tooltip_content } from './tooltip-content-template';
import { mathMin, mathMax, mathAbs } from '../core/zc-constants';
import { d3_ease_linear, d3_range, d3_event, d3_interpolateRound } from '../core/d3-constants';
import { $Browser } from '../core/browser-detect';
import { getDashArray } from '../utils/dashstyles';
import { tooltip_helpers } from './tooltip-helpers';
import { $ZCG } from '../core/zcglobal';
import { TRANSPARENT, NULL } from '../core/zc-constants';
import { getSVGImage } from '../effects/dataImage';
import { SANKEY_CHARTNAME } from '../core/chartname-constants';
import { getChartTypeByIndex, isAlluvialChartType, isMultiColorChart } from '../core/charttype-helpers';
import { SELECTED_MERGED } from '../events/dataSelection/selection-types';

events.tooltip = function(element, data, chartObj) {
    'use strict';
    var MAX_WIDTH = '40%',
        MAX_HEIGHT = '100%';

    var tooltip = {
            element: element
        },
        systemConf = chartObj.systemConf,
        dataset = chartObj.dataset,
        dataSplicer = chartObj.dataSplicer,
        metadata_axes = dataSplicer.getEncode(),
        metadata_columns = dataSplicer.getColumns(),
        tooltipIndices = metadata_axes.tooltip || d3_range(metadata_columns.length),
        tooltip_data = systemConf.tooltip,
        themeMaxWidth = null,
        tooltipPosition = tooltip_data.position || 'relative',
        hookEnabled = defined(tooltip_data.hook) ? tooltip_data.hook : tooltipPosition === 'relative',
        hookPosition = {
            left: 50,
            top: 100,
            rotation: 0
        };

    if (tooltip_data.theme && tooltip_data.theme !== '') {
        var theme = isObjectPropDefined(window.$ZC, 'tooltip.getTheme') && $ZC.tooltip.getTheme(tooltip_data.theme, chartObj);
        if (theme !== null) {
            tooltip_data = chartObj.systemConf.tooltip = mergeJSON(tooltip_data, theme.borderStyle, true, true);
            chartObj.tooltip.content = [theme.body];
            themeMaxWidth = theme.maxWidth;
        }
    }

    var hookHeight = validateNullVal(tooltip_data.hookHeight, 6),
        dist = validateNullVal(tooltip_data.outerPadding, 20),
        borderWidth = defined(tooltip_data.borderWidth) ? tooltip_data.borderWidth : 1,
        borderRadius = defined(tooltip_data.borderRadius) ? tooltip_data.borderRadius : 5,
        borderStyle = defined(tooltip_data.borderStyle) ? tooltip_data.borderStyle : 'solid',
        padding = defined(tooltip_data.padding) ? `${tooltip_data.padding}` : `${5}`,
        isWithinChart = validateNullVal(tooltip_data.withinChart, true),
        hookBorderWidth = validateNullVal(tooltip_data.hookBorderWidth, / /g.test(borderWidth) ? 1 : +borderWidth),
        parsedBorderRadius = getBorderRadius(borderRadius);

    var maxWidth = perToPx(tooltip_data.maxWidth || themeMaxWidth || MAX_WIDTH, chartObj.chartarea.width),
        maxHeight = perToPx(tooltip_data.maxHeight || MAX_HEIGHT, chartObj.chartarea.height),
        minWidth = mathMax(chartObj.chartarea.width * 0.05, 2 * hookHeight);

    var headerContentFunc = tooltip_data.headerContent,
        footerContentFunc = tooltip_data.footerContent,
        crosshair_data = tooltip_data.crosshair;

    // Tooltip elements
    var arrowbefore, arrowafter, arrowBeforeNAfter, contentBox, backgroundElement, effectOnChild;

    tooltip.initialize = function() {
        var fontSize = getFontSizeByHierarchy(tooltip_data, chartObj),
            defaultLineHeight = +fontSize + +fontSize * 0.35,
            userdataCanvasFontColor = chartObj.userdata.canvas != null ? chartObj.userdata.canvas.fontColor : null,
            borderRadius_px = shorthand_px(borderRadius),
            filterDisabled = tooltip_data.useChartEffect === false, // enabled default
            filterEffect = null;
        effectOnChild = false;

        if (!filterDisabled && chartObj.filterEffectsID) {
            filterEffect = `url(#${chartObj.filterEffectsID})`;
            effectOnChild = chartObj.systemConf.chart.effects.choice !== 'dropShadow' && !$Browser.PHANTOMJS;
        }

        element.styles({
            width: 'auto',
            opacity: tooltip_data.opacity,
            'z-index': tooltip_data.zindex,
            '-webkit-transform-style': 'preserve-3d',
            '-moz-transform-style': 'preserve-3d',
            '-ms-transform-style': 'preserve-3d',
            '-o-transform-style': 'preserve-3d',
            'transform-style': 'preserve-3d',
            // "-webkit-backface-visibility": "hidden",
            // "-moz-backface-visibility": "hidden",
            // "-ms-backface-visibility": "hidden",
            // "-o-backface-visibility": "hidden",
            // "backface-visibility": "hidden",
            'pointer-events': tooltip_data.pointerEvents
            // "padding": shorthand_px(padding)
        });

        backgroundElement = appendEle(element, 'div', effectOnChild ? [1] : [], '.', 'class', 'zc-tooltip-background');

        if (effectOnChild) {
            element.styles({
                'box-shadow': null,
                background: null,
                border: 'none',
                filter: null
            });
            backgroundElement
                .styles({
                    position: 'absolute',
                    top: 0,
                    left: 0,
                    width: '100%',
                    height: '100%',
                    'border-radius': borderRadius_px,
                    'border-width': shorthand_px(borderWidth),
                    'border-style': borderStyle,
                    'box-shadow': tooltip_data.shadow,
                    filter: filterEffect,
                    'z-index': -1
                })
                .lower();
            element.selectAll('div#zctooltipbefore,div#zctooltipafter').remove();
        } else {
            element.styles({
                'box-shadow': tooltip_data.shadow,
                'border-radius': borderRadius_px,
                'border-width': shorthand_px(borderWidth),
                'border-style': borderStyle,
                filter: filterEffect
            });
        }

        var borderTopBox = appendEle(element, 'div', tooltip_data.borderTop ? [tooltip_data.borderTop] : [], '.', 'class', 'borderTop');
        borderTopBox.styles({
            height: `${tooltip_data.borderTop}px`,
            'border-radius': borderRadius_px,
            'border-bottom-left-radius': '1px',
            'border-bottom-right-radius': '1px'
        });

        contentBox = appendEle(element, 'div', [1], '#', 'id', 'tooltipcontent');
        contentBox.styles({
            'box-sizing': tooltipPosition === 'relative' ? null : 'border-box',
            'max-width': tooltipPosition === 'relative' ? `${maxWidth}px` : null,
            'min-width': tooltipPosition === 'relative' ? `${minWidth}px` : null,
            // "max-height" : maxHeight + "px", // Is it necessary?
            'line-height': `${tooltip_data.lineHeight || defaultLineHeight}px`,
            'word-wrap': 'break-word',
            'white-space': 'normal',
            'overflow-y': 'auto',
            padding: shorthand_px(padding)
            //		"border-top":"1px solid red"
        });
        applyFont(
            contentBox,
            [
                {
                    fontColor:
                        tooltip_data.fontColor == null && userdataCanvasFontColor == null
                            ? 'white'
                            : tooltip_data.fontColor || userdataCanvasFontColor
                },
                tooltip_data,
                chartObj.systemConf.canvas,
                $ZCG
            ],
            true
        );

        // arrow before
        arrowbefore = appendEle(effectOnChild ? backgroundElement : element, 'div', hookEnabled ? [1] : [], '#', 'id', 'zctooltipbefore');
        arrowbefore.styles({
            'margin-left': `${-(hookHeight + 2 * hookBorderWidth)}px`,
            border: `${hookHeight + 2 * hookBorderWidth}px solid transparent`,
            filter: effectOnChild ? filterEffect : null
        });

        // arrow after
        arrowafter = appendEle(effectOnChild ? backgroundElement : element, 'div', hookEnabled ? [1] : [], '#', 'id', 'zctooltipafter');
        arrowafter.styles({
            'margin-left': `${-hookHeight}px`,
            border: `${hookHeight}px solid transparent`,
            filter: effectOnChild ? filterEffect : null
        });

        // Properties for arrow before and after
        arrowBeforeNAfter = element.selectAll('div#zctooltipbefore,div#zctooltipafter');
        arrowBeforeNAfter.styles({
            'border-color': TRANSPARENT,
            'border-top-color': 'white',
            position: 'absolute',
            top: '100%',
            left: '50%',
            width: 0,
            height: 0,
            content: ' '
        });
    };

    tooltip.show = function(chartObj, position, content, useDefault, defaultArgs) {
        element = chartObj.tooltip.element;
        arrowBeforeNAfter = element.selectAll('div#zctooltipbefore,div#zctooltipafter');
        if (!defined(defaultArgs)) {
            defaultArgs = {};
        }

        var color = defaultArgs.color,
            svgFillFunc = defaultArgs.svgFillFunc,
            seriesindex = defaultArgs.seriesindex,
            subseriespos = defaultArgs.subseriespos,
            subseriesLength = defaultArgs.subseriesLength,
            itempos = defaultArgs.itempos,
            paneid = defaultArgs.paneid,
            backgroundColor = defaultArgs.backgroundColor || tooltip_data.backgroundColor,
            borderColor = defaultArgs.borderColor || tooltip_data.borderColor,
            tooltipSharedInfo = defaultArgs.tooltipSharedInfo,
            tooltipShared = defaultArgs.tooltipShared,
            cType =
                (seriesindex != null && getChartTypeByIndex(chartObj, paneid, seriesindex).name) ||
                (defaultArgs.chartType && defaultArgs.chartType.name),
            hasMultiColoring,
            renderTo = defaultArgs.renderTo || 'plot', //Bydefault renderTo is 'plot', for legend selectAllbox it is 'legendTitle'
            hideHook;
        useDefault = content == null && useDefault === true;

        const isLevelAxis = isAlluvialChartType(chartObj);

        // getting color
        if (useDefault || defaultArgs.inheritColors) {
            hasMultiColoring = isMultiColorChart(cType, chartObj);
        }
        var clr = color;
        // var legendOrder = chartObj.dataObject.pivot && paneid ? chartObj.dataObject.hierarchical.gridOptions[paneid].legendOrder || 0 : null;
        if (useDefault || defaultArgs.inheritColors) {
            var idx = hasMultiColoring ? itempos : seriesindex;
            let colidx = isLevelAxis ? dataset.getColumnIdx('level', getDepth(defaultArgs.d.level) - 1) : null;
            var dataPoint = dataset.getDataPoint(chartObj, colidx, seriesindex, subseriespos, itempos, defaultArgs.d);
            clr = getRawColor(chartObj, idx, dataPoint, seriesindex, paneid, dataPoint && dataPoint.rendererIndex);
        } else if (isArray(clr)) {
            clr = clr[seriesindex % clr.length];
        }

        if (tooltipShared) {
            if (backgroundColor == 'inherit') {
                backgroundColor = defaultArgs.nearest ? clr : 'black';
            }
            if (borderColor == 'inherit') {
                borderColor = defaultArgs.nearest ? clr : 'black';
            }
            clr = 'black';
        } else {
            if (backgroundColor == 'inherit') {
                backgroundColor = clr;
            }
            if (borderColor == 'inherit') {
                borderColor = clr;
            }
        }
        borderColor = borderColor != null ? borderColor : clr;
        backgroundColor = backgroundColor != null ? backgroundColor : clr;

        element.style('visibility', 'visible');
        (effectOnChild ? backgroundElement : element).styles({
            'background-color': backgroundColor,
            'border-color': borderColor
        });
        var selectionManager = chartObj.eventHandler.eventManager.selectionManager,
            nearest = selectionManager.nearest;
            //For analytics renrendering image case
            //For boxplot custom tooltip indices case
        if (!nearest || (nearest && nearest.selected != SELECTED_MERGED)  || (typeof tooltipIndices == 'function' || typeof tooltipIndices == 'string' )) {
            //***** TOOLTIP INNERHTML
            var tooltipcontent = '',
                toUserInfo,
                headerContent = '',
                footerContent = '',
                temp,
                contentBoxNode = contentBox.node();

            //custom header content
            if (useDefault && headerContentFunc != null) {
                headerContent = invokeFunction(headerContentFunc, d3_event(), (toUserInfo = getInfoToUser()), chartObj);
            }

            //content
            if (useDefault) {
                if (tooltip_data.content) {
                    var defaultContent = function(){
                        return getDefaultContent(seriesindex, subseriespos, itempos, clr, defaultArgs.d);
                    }
                    tooltipcontent += invokeFunction(
                        tooltip_data.content,
                        d3_event(),
                        (toUserInfo = toUserInfo || getInfoToUser()),
                        chartObj,
                        defaultContent
                    );
                } else if (tooltipShared) {
                    tooltipcontent += getTooltipSharedContent(
                        seriesindex,
                        subseriespos,
                        itempos,
                        defaultArgs.d,
                        tooltipSharedInfo,
                        color,
                        hasMultiColoring,
                        subseriesLength
                    );
                } else {
                    tooltipcontent += getDefaultContent(seriesindex, subseriespos, itempos, clr, defaultArgs.d);
                }
            } else {
                tooltipcontent += content;
            }

            //custom footer content
            if (useDefault && footerContentFunc != null) {
                footerContent = invokeFunction(footerContentFunc, d3_event(), (toUserInfo = toUserInfo || getInfoToUser()), chartObj);
            }

            // add html first
            contentBox.html(tooltipcontent);

            if (!isHTMLElement(headerContent)) {
                temp = document.createElement('div');
                temp.innerHTML = headerContent;
                headerContent = temp;
            }

            // add add header first
            if (!isHTMLElement(footerContent)) {
                temp = document.createElement('div');
                temp.innerHTML = footerContent;
                footerContent = temp;
            }

            // add add header to the top
            contentBoxNode.insertBefore(headerContent, contentBoxNode.firstChild);
            contentBoxNode.appendChild(footerContent);

            // TOOLTIP INNERHTML *****
        }
        var elementNode = element.node(),
            eleWidth = elementNode.offsetWidth,
            eleHeight = elementNode.offsetHeight;

        if (renderTo == 'plot' && crosshair_data.type != NONE && chartObj.isAxisCategory) {
            showCrossshair(chartObj, position.actLeft || position.left, position.actTop || position.top, crosshair_data);
        }

        if (renderTo == 'plot' && position.helpers && !position.helpers.isWithinPlot) {
            hideHook = true;
        }

        element.interrupt().transition();

        element.transition().delay(0);

        var coordinate;
        if (tooltipPosition === 'relative') {
            var curx = parseFloat(elementNode.style.left),
                cury = parseFloat(elementNode.style.top);
            coordinate = $ZC_TOOLTIP_POSITION.relative(
                position,
                element,
                chartObj,
                hookHeight,
                mathAbs(dist),
                isWithinChart,
                hookBorderWidth,
                parsedBorderRadius,
                maxWidth,
                maxHeight
            );
            hookPosition = coordinate.hookPosition;

            if (!curx || isNaN(curx) || !cury || isNaN(cury)) {
                // Transition will start after ~4ms only. If the tooltip is called at the first time no transition
                element.styles({
                    left: `${coordinate.x}px`,
                    top: `${coordinate.y}px`
                });
            }

            curx = !curx || isNaN(curx) ? coordinate.x : (2 * curx + coordinate.x) / 3;
            cury = !cury || isNaN(cury) ? coordinate.y : (cury + coordinate.y) / 2;

            element
                .transition()
                .duration(100)
                .ease(d3_ease_linear)
                .styleTween('left', positionTween(curx, coordinate.x)) //To avoid sub pixel rendering issue, d3.interpolateRound has been used
                .styleTween('top', positionTween(cury, coordinate.y));
        } else if (tooltipPosition === 'fixed') {
            coordinate = $ZC_TOOLTIP_POSITION.fixed(position, element, chartObj);
            element.styles({
                left: `${coordinate.x}px`,
                top: `${coordinate.y}px`
            });
        } else {
            coordinate = $ZC_TOOLTIP_POSITION.block(position, element, chartObj);
            element
                .styles({
                    left: `${coordinate.x}px`,
                    top: `${coordinate.y}px`
                })
                .attr('align', coordinate.align); // test in all browsers
        }

        contentBox.styles({
            width: coordinate.w,
            height: coordinate.h,
            'max-width': coordinate.maxw,
            'max-height': coordinate.maxh
        });

        //borderTop
        if (tooltip_data.borderTop) {
            var svgFillEle = svgFillFunc
                ? svgFillFunc(
                      hasMultiColoring ? itempos : seriesindex,
                      null,
                      null,
                      clr,
                      null,
                      null,
                      { width: mathMin(+coordinate.w, +coordinate.maxw) },
                      true
                  )
                : borderColor;
            var svgFillUrl = typeof svgFillEle === 'object' ? `url(${getSVGImage(svgFillEle, true).url})` : svgFillEle;
            var borderTop = element.selectAll('div.borderTop');
            borderTop.styles({
                background: svgFillUrl
            });
        }

        // Tooltip arrow
        if (tooltipShared || hideHook) {
            arrowBeforeNAfter.style('display', NONE);
        } else {
            var arrowTransform = `rotate(${hookPosition.rotation}deg)`,
                arrowTransformOrigin = '50% 0%';

            arrowBeforeNAfter.styles({
                display: '',
                left: `${hookPosition.left}%`,
                top: `${hookPosition.top}%`,
                transform: arrowTransform,
                '-webkit-transform': arrowTransform,
                '-ms-transform': arrowTransform,
                'transform-origin': arrowTransformOrigin,
                '-webkit-transform-origin': arrowTransformOrigin,
                '-ms-transform-origin': arrowTransformOrigin
            });

            arrowbefore.style('border-top-color', borderColor);

            arrowafter.styles({
                'border-top-color': backgroundColor,
                opacity: 1
            });
        }

        function getInfoToUser() {
            if (tooltipShared) {
                return tooltipSharedInfo.keys().map(function(d) {
                    var curSerpos = pInt(d),
                        curItempos = pInt(tooltipSharedInfo.get(d)),
                        clridx = dataset.getColumnIdx('clr');
                    var yaxiscolumnorder = chartObj.seriesdata[curSerpos].yaxiscolumnorder || [0, 0];
                    var validx = chartObj.ismultiy || clridx == null ? dataset.getColumnIdx('y', yaxiscolumnorder) : clridx;
                    var dp = dataset.getDataPoint(chartObj, validx, curSerpos, subseriespos, curItempos);
                    var metaInfoObj = dataset.getMetaInfoObj(chartObj, dp, curSerpos);
                    return {
                        actData: dp,
                        metaInfo: metaInfoObj
                    };
                });
            }
            var dp = dataset.getDataPoint(chartObj, null, seriesindex, subseriespos, itempos, defaultArgs.d); // why we are sending null column index
            var metaInfoObj = chartObj.dataset.getMetaInfoObj(chartObj, dp, seriesindex);
            return {
                actData: dp,
                point: defaultArgs.d,
                metaInfo: metaInfoObj,
                seriesIndex: seriesindex,
                subseriesIndex: subseriespos,
                itemIndex: itempos
            };
        }
    };

    tooltip.hide = function(tooltipele) {
        tooltipele = tooltipele == null ? this.element : tooltipele;
        tooltipele.style('visibility', 'hidden');
        if (chartObj.systemConf.tooltip.crosshair.type != NONE) {
            chartObj.staticPaneSVG.selectAll('g.crosshair').style('visibility', 'hidden');
        }
    };

    tooltip.destroy = function() {
        this.element && this.element.remove();
        this.element = arrowbefore = arrowafter = null;     // memory leak fix
    };

    tooltip.getTooltipContent = function() {
        /* arguments
         * 0 - seriesIndex
         * 1 - subseriesIndex
         * 2 - itemIndex
         */

        return getDefaultContent.apply(null, arguments);
    };

    function showCrossshair(chartObj, x, y, data) {
        chartObj.staticPaneSVG.selectAll('g.crosshair').style('visibility', 'visible');

        var type = data.type;
        if (type == 'x' || type == 'xy') {
            drawCrosshairLine(chartObj, x, y, data, 'x');
        }
        if (type == 'y' || type == 'xy') {
            drawCrosshairLine(chartObj, x, y, data, 'y');
        }
    }

    function drawCrosshairLine(chartObj, x, y, data, type) {
        var ele = chartObj.staticPaneSVG.selectAll(`line.${type}-crosshair`);
        var isVerticalLine = (type == 'x' && !chartObj.axes.rotated) || (type == 'y' && chartObj.axes.rotated);
        ele.attrs({
            x1: isVerticalLine ? x : chartObj.plotarea.left,
            y1: isVerticalLine ? chartObj.plotarea.top : y,
            x2: isVerticalLine ? x : chartObj.plotarea.left + chartObj.plotarea.width,
            y2: isVerticalLine ? chartObj.plotarea.height + chartObj.plotarea.top : y
        }).styles({
            'shape-rendering': 'crispEdges',
            stroke: data.color,
            'stroke-width': data.strokeWidth,
            'stroke-dasharray': getDashArray(data.dashStyle, data.strokeWidth),
            'pointer-events': NONE
        });
    }

    function getDefaultContent(seriesindex, subseriespos, itempos, color, d, customTooltipIndices) {
        if (isObjectPropDefined(window.$ZC, 'tooltip.getTheme') && defined(systemConf.tooltip.theme) && systemConf.tooltip.theme !== '') {
            if (defined(chartObj.tooltip.content)) {
                return dataset.getCustomizedVal(
                    chartObj,
                    chartObj.tooltip.content,
                    seriesindex,
                    subseriespos,
                    itempos,
                    d,
                    'tooltip',
                    color
                );
            }
        }

        var chartName = getChartTypeByIndex(chartObj, d && d.paneid, seriesindex).name,
            hasMultiColoring = isMultiColorChart(chartName, chartObj),
            dataPoint = dataset.getDataPoint(chartObj, null, seriesindex, subseriespos, itempos, d),
            clr = getRawColor(chartObj, hasMultiColoring ? itempos : seriesindex, dataPoint, seriesindex),
            indices_ = invokeFunction(customTooltipIndices || tooltipIndices, null, dataPoint, chartObj),
            indices = indices_.filter(function(idx) {
                return !(metadata_columns[idx] && isUndefined(dataset.getDataPoint(chartObj, idx, seriesindex, subseriespos, itempos, d))); // dataPoint is null for the current idx
            }),
            isSankeyChart = chartName === SANKEY_CHARTNAME,
            template;

        // if((chartName == BAR_CHARTNAME || chartName == BARRANGE_CHARTNAME) && chartObj.dataObject.hasWhiskers){
        //
        // }

        if (isSankeyChart) {
            if (chartObj.isLevelAxisCategory) {
                template = tooltip_content.getContentTemplate(chartObj, NULL, indices);
                d = d.original || d;
            } else if (d._type === 'node') {
                let xIndex = chartObj.dataset.xidx,
                zColIdx = chartObj.dataset.getColumnIdx('z'),
                xColIdx = chartObj.dataset.getColumnIdx(d._format || 'x', [0, d._axisOrder]);
                template = `{{col(${xColIdx})}} : {{val(${xIndex})}}<br>{{col(${zColIdx})}} : {{val(${zColIdx})}}`;
            } else {
                template = tooltip_content.getContentTemplate(chartObj, NULL, indices);
            }
        } else {
            template = tooltip_content.getContentTemplate(chartObj, NULL, indices);
        }

        return dataset.getCustomizedVal(chartObj, template, seriesindex, subseriespos, itempos, d, 'tooltip', clr);
    }

    function getTooltipSharedContent(seriesindex, subseriespos, itempos, d, tooltipSharedInfo, color, hasMultiColoring, subseriesLength) {
        var clridx = dataset.getColumnIdx('clr'),
            yaxiscolumnorder_series = chartObj.seriesdata[seriesindex].yaxiscolumnorder || [0, 0], //yaxiscolumnorder is not mandatory
            repeatIndex =
                chartObj.dataObject.pieWithAxis && !chartObj.dataObject.isCombinationalChart
                    ? tooltip_helpers.getPieColumnIndex(chartObj)
                    : dataset.getColumnIdx('y', yaxiscolumnorder_series),
            sharedKeys = tooltipSharedInfo.keys(),
            y_pattern = tooltipSharedInfo.y_val_pattern,
            indices_ = invokeFunction(tooltipIndices || tooltipIndices, null, d),
            meta_tooltip = tooltip_content.getRawSharedContent(chartObj, tooltipSharedInfo, indices_),
            i = 0,
            updatePattern = function(pattern, curSerpos, subseriespos, curItempos, clr, copy_y_pattern) {
                if (chartObj.ismultiy || clridx == null) {
                    var seriesName = encodeHTML(chartObj.seriesdata[curSerpos].seriesname) || copy_y_pattern.replace(/val/gi, 'col');
                    pattern = pattern.replace('{{val(null)}}', seriesName);
                }
                return dataset.getCustomizedVal(
                    chartObj,
                    pattern,
                    curSerpos,
                    subseriespos,
                    curItempos,
                    NULL, //Doubt for unsorted seriesdataafterupdate
                    'tooltip',
                    clr,
                    NULL,
                    subseriesLength && subseriesLength > 1 ? subseriesLength : NULL
                );
            },
            removeIndices = [];

        meta_tooltip.forEach(function(pattern, patternIndex) {
            var hasRepeatPattern = (isArray(pattern) ? pattern : [pattern]).some(function(cur) {
                return (
                    typeof cur === 'string' &&
                    (cur.indexOf(`{{val(${repeatIndex})}}`) != -1 || cur.indexOf(`{{per(${repeatIndex})}}`) != -1)
                );
            });

            if (isArray(pattern) || hasRepeatPattern) {
                i = i % chartObj.seriesdata.length; //#ZC1248
                var curKey = sharedKeys[i];

                if (hasRepeatPattern) {
                    i++;
                }

                //y-values
                if (isUndefined(curKey)) {
                    // If datapoint is not available
                    removeIndices.push(patternIndex);
                    return;
                }
                var curSerpos = pInt(curKey),
                    curItempos = pInt(tooltipSharedInfo.get(curKey)),
                    copy_y_pattern = y_pattern,
                    clr = color;

                if (isArray(clr)) {
                    var dataPoint = dataset.getDataPoint(chartObj, null, curSerpos, subseriespos, curItempos, d);
                    clr = getRawColor(chartObj, hasMultiColoring ? curItempos : curSerpos, dataPoint, curSerpos);
                }

                if (isArray(pattern)) {
                    pattern.forEach(function(cur, j) {
                        pattern[j] = updatePattern(pattern[j], curSerpos, subseriespos, curItempos, clr, copy_y_pattern);
                    });
                } else {
                    meta_tooltip[patternIndex] = updatePattern(
                        meta_tooltip[patternIndex],
                        curSerpos,
                        subseriespos,
                        curItempos,
                        clr,
                        copy_y_pattern
                    );
                }
            } else {
                //Custom text with non y-values
                var curKey = sharedKeys[0],
                    curSerpos,
                    curItempos;

                if (defined(curKey)) {
                    curSerpos = pInt(curKey);
                    curItempos = pInt(tooltipSharedInfo.get(curKey));
                } else {
                    curSerpos = seriesindex;
                    curItempos = itempos;
                }
                meta_tooltip[patternIndex] = dataset.getCustomizedVal(
                    chartObj,
                    pattern,
                    curSerpos,
                    subseriespos,
                    curItempos,
                    d,
                    'tooltip',
                    color,
                    NULL
                );
            }
        });

        // If datapoint is not available
        meta_tooltip = meta_tooltip.filter(function(d, i) {
            return removeIndices.indexOf(i) == -1;
        });

        return tooltip_content.getContentTemplate(chartObj, meta_tooltip);
    }

    function positionTween(start, end) {
        return function(d, i) {
            var interpolate = d3_interpolateRound(parseInt(start), parseInt(end));
            return function(t) {
                return `${interpolate(t)}px`;
            };
        };
    }

    return tooltip;
};

export var _chartTooltip = events.tooltip;
Registry.setComponent('tooltip', _chartTooltip);
