// @zohocharts/zc/src/tooltip/tooltip-position.js
//$Id$
//tooltip-position.js
import { tooltip_helpers } from './tooltip-helpers';
// import { tooltip_block_helpers } from './tooltip-blockPosition';
import { mathAbs, mathSQRT, mathMin, PX, mathMax, mathPow } from '../core/zc-constants';
import { TOP_ALIGN, RIGHT_ALIGN, LEFT_ALIGN, BOTTOM_ALIGN, CENTER_ALIGN } from '../core/align-constants';
import { pick, parseShortHandValue, shorthand_px, defined, isUndefined, perToPx } from '../utils/zcutils';
import { tooltip } from '../core/main-zc-init';

export var $ZC_TOOLTIP_POSITION = {};
$ZC_TOOLTIP_POSITION.relative = function(
    position,
    element,
    chartObj,
    hookHeight,
    dist,
    isWithinChart,
    hookBorderWidth,
    parsedBorderRadius,
    defaultMaxWidth,
    defaultMaxHeight
) {
    element.selectAll('#tooltipcontent').styles({
        width: null, //"auto",
        height: null,
        'max-width': defaultMaxWidth + PX,
        'max-height': defaultMaxHeight + PX
    });

    var eleNode = element.node(),
        eleWidth = eleNode.offsetWidth,
        eleHeight = eleNode.offsetHeight,
        marginRight = 2,
        canvasarea = chartObj.canvasarea,
        chartarea = chartObj.chartarea,
        distance = dist + hookHeight + (position.distance || 0),
        dTop = position.dTop || 0,
        dLeft = position.dLeft || 0,
        tooltipConfig = chartObj.systemConf.tooltip,
        orient = (tooltipConfig.relative || {}).orient,
        borderTop = tooltipConfig.borderTop || 0,
        availableWidth = chartarea.width - chartarea.left + chartarea.right,
        availableHeight = chartarea.height,
        dim = tooltip_helpers.getSize(
            tooltipConfig,
            [availableWidth, availableHeight],
            [eleWidth, eleHeight],
            [defaultMaxWidth, defaultMaxHeight]
        ),
        width = dim.w,
        height = dim.h,
        hookPosition = {
            left: 50,
            top: 100,
            rotation: 0
        },
        checkBounds = {
            left: mathAbs(mathMin(dLeft, 0)) + (chartObj.scrolledX || 0),
            top: mathAbs(mathMin(dTop, 0)) + (chartObj.scrolledY || 0),
            right: canvasarea.width - marginRight - mathMax(dLeft, 0) + mathAbs(mathMin(dLeft, 0)), // width + left
            bottom: canvasarea.height - mathMax(dTop, 0) + mathAbs(mathMin(dTop, 0)) // height + top
        },
        coordinate = {
            x: position.left - width / 2,
            y: position.top - height - distance,
            // w: width + PX,
            // h: (height - borderTop) + PX,
            maxw: defaultMaxWidth + PX,
            maxh: defaultMaxHeight + PX
        };

    if (!(!isWithinChart || chartObj.scrolledX || chartObj.scrolledY)) {
        var maxTop = position.top - height - distance,
            maxRight = position.left + width + distance,
            maxLeft = position.left - width - distance,
            maxBottom = position.top + height + distance;

        orient =
            orient ||
            (maxTop > checkBounds.top
                ? TOP_ALIGN
                : maxRight < checkBounds.right
                ? RIGHT_ALIGN
                : maxLeft > checkBounds.left
                ? LEFT_ALIGN
                : maxBottom < checkBounds.bottom
                ? BOTTOM_ALIGN
                : null);
        if (orient == TOP_ALIGN) {
            // TOP
            positionHook(TOP_ALIGN);
            hookPosition.rotation = 0;
            hookPosition.top = 100;
        } else if (orient == RIGHT_ALIGN) {
            // RIGHT
            coordinate.x = position.left + distance;
            positionHook(RIGHT_ALIGN);
            hookPosition.rotation = 90;
            hookPosition.left = 0;
        } else if (orient == LEFT_ALIGN) {
            // LEFT
            coordinate.x = position.left - distance - width;
            positionHook(LEFT_ALIGN);
            hookPosition.rotation = -90;
            hookPosition.left = 100;
        } else if (orient == BOTTOM_ALIGN) {
            // BOTTOM
            positionHook(BOTTOM_ALIGN);
            coordinate.y = position.top + distance;
            hookPosition.rotation = 180;
            hookPosition.top = 0;
        }
    }
    coordinate.hookPosition = hookPosition;
    return coordinate;

    function positionHook(mode) {
        var rad;
        if (/(top|bottom)/.test(mode)) {
            if (position.left - width / 2 < checkBounds.left) {
                // Is within left?
                rad = getRad(mode == TOP_ALIGN ? 'blx' : 'tlx');
                coordinate.x = checkBounds.left;
                hookPosition.left = (mathMax(rad, position.left - checkBounds.left) / width) * 100;
            } else if (position.left + width / 2 > checkBounds.right) {
                // Is within right?
                rad = getRad(mode == TOP_ALIGN ? 'brx' : 'trx');
                coordinate.x = checkBounds.right - width;
                hookPosition.left = (1 - mathMax(rad, checkBounds.right - position.left) / width) * 100;
            } else {
                hookPosition.left = 50;
            }
        } else {
            if (position.top - height / 2 < checkBounds.top) {
                // Is within top?
                rad = getRad(mode == LEFT_ALIGN ? 'trY' : 'tly');
                coordinate.y = checkBounds.top;
                hookPosition.top = (mathMax(rad, position.top - checkBounds.top) / height) * 100;
            } else if (position.top + height / 2 > checkBounds.bottom) {
                // Is within bottom?
                rad = getRad(mode == LEFT_ALIGN ? 'bry' : 'bly');
                coordinate.y = checkBounds.bottom - height;
                hookPosition.top = (1 - mathMax(rad, checkBounds.bottom - position.top) / height) * 100;
            } else {
                coordinate.y = position.top - height / 2;
                hookPosition.top = 50;
            }
        }
    }

    function getRad(attr) {
        return hookHeight + parsedBorderRadius[attr] + mathSQRT(mathPow(hookBorderWidth, 2) + mathPow(hookBorderWidth, 2));
    }
};

$ZC_TOOLTIP_POSITION.fixed = function(position, element, chartObj) {
    element.selectAll('#tooltipcontent').styles({
        width: null, //"auto",
        height: null,
        'max-width': null,
        'max-height': null
    });
    element.style('border-width', 0);
    var elementNode = element.node();

    var eleWidth = elementNode.offsetWidth,
        eleHeight = elementNode.offsetHeight,
        canvasarea = chartObj.canvasarea,
        dTop = position.dTop || 0,
        dLeft = position.dLeft || 0,
        tooltipConfig = chartObj.systemConf.tooltip,
        orient = (tooltipConfig.fixed || {}).orient,
        borderTop = tooltipConfig.borderTop || 0,
        borderWidth = pick(tooltipConfig.borderWidth, 1),
        parsedBorderWidth = parseShortHandValue(borderWidth),
        borderWidthTop = parsedBorderWidth.top,
        borderWidthRight = parsedBorderWidth.right,
        borderWidthBottom = parsedBorderWidth.bottom,
        borderWidthLeft = parsedBorderWidth.left,
        availableWidth = canvasarea.width - borderWidthLeft - borderWidthRight,
        availableHeight = canvasarea.height - borderWidthTop - borderWidthBottom,
        dim = tooltip_helpers.getSize(tooltipConfig, [availableWidth, availableHeight], [eleWidth, eleHeight]),
        width = dim.w,
        height = dim.h,
        checkBounds = {
            left: 0,
            top: 0,
            right: window.innerWidth, // width + left
            bottom: window.innerHeight // height + top
        },
        coordinate = {
            x: mathAbs(mathMin(dLeft, 0)),
            y: mathAbs(mathMin(dTop, 0))
        },
        x = mathAbs(mathMin(dLeft, 0)),
        y = mathAbs(mathMin(dTop, 0)),
        w = width,
        h = height,
        maxh,
        maxTop = coordinate.y - height,
        maxRight = coordinate.x + canvasarea.width + width,
        maxLeft = coordinate.x - width,
        maxBottom = coordinate.y + canvasarea.height + height,
        hasEnoughHeightAfterScrollY = dTop <= 0 || (dTop > 0 && canvasarea.height - dTop > height);

    if (!orient) {
        if (maxTop > checkBounds.top) {
            orient = TOP_ALIGN;
        } else if (maxRight < checkBounds.right && hasEnoughHeightAfterScrollY) {
            orient = RIGHT_ALIGN;
        } else if (maxLeft > checkBounds.left && hasEnoughHeightAfterScrollY) {
            orient = LEFT_ALIGN;
        } else if (maxBottom < checkBounds.bottom) {
            orient = BOTTOM_ALIGN;
        }
    }

    if (orient === TOP_ALIGN) {
        coordinate.y = y - height - borderWidthTop - borderWidthBottom;
        w = availableWidth - (dLeft > 0 ? dLeft : 0);
        maxh = '40%';
    } else if (orient === RIGHT_ALIGN) {
        coordinate.x = x + canvasarea.width;
        h = availableHeight - (dTop > 0 ? dTop : 0);
    } else if (orient === LEFT_ALIGN) {
        coordinate.x = x - width - borderWidthRight - borderWidthLeft;
        h = availableHeight - (dTop > 0 ? dTop : 0);
    } else if (orient === BOTTOM_ALIGN) {
        coordinate.y = y + canvasarea.height - (dTop > 0 ? dTop : 0);
        w = availableWidth - (dLeft > 0 ? dLeft : 0);
        maxh = '40%';
    } else {
        coordinate.y = y;
        w = availableWidth - (dLeft > 0 ? dLeft : 0);
        maxh = '40%';
    }

    dim = tooltip_helpers.getSize(tooltipConfig, [availableWidth, availableHeight], [w, h]);
    coordinate.w = dim.w + PX;
    coordinate.h = dim.h - borderTop + PX;
    coordinate.maxw = availableWidth - (dLeft > 0 ? dLeft : 0) + PX;
    coordinate.maxh =
        mathMin(
            availableHeight - (dTop > 0 ? dTop : 0),
            isUndefined(tooltipConfig.height) && isUndefined(tooltipConfig.maxHeight) && defined(maxh)
                ? perToPx(maxh, availableHeight)
                : Infinity
        ) + PX;

    element.style('border-width', shorthand_px(borderWidth));
    return coordinate;
};

$ZC_TOOLTIP_POSITION.block = function(position, element, chartObj) {
    var { chartarea, plotarea } = chartObj,
        tooltipConfig = chartObj.systemConf.tooltip,
        blockConfig = tooltipConfig.block || {},
        borderTop = tooltipConfig.borderTop || 0,
        borderWidth = pick(tooltipConfig.borderWidth, 1),
        parsedBorderWidth = parseShortHandValue(borderWidth),
        borderWidthTop = parsedBorderWidth.top,
        borderWidthRight = parsedBorderWidth.right,
        borderWidthBottom = parsedBorderWidth.bottom,
        borderWidthLeft = parsedBorderWidth.left,
        vAlign = blockConfig.vAlign || TOP_ALIGN,
        hAlign = blockConfig.hAlign,
        dTop = position.dTop || 0,
        dLeft = position.dLeft || 0,
        availableWidth = chartarea.width - borderWidthLeft - borderWidthRight,
        availableHeight = chartarea.height,
        dim = tooltip_helpers.getSize(tooltipConfig, [availableWidth, availableHeight], ['100%', chartObj.tooltip.preCalculatedHeight]),
        w = dim.w,
        h = mathMin(dim.h, perToPx(tooltip.block_helpers.MAXHEIGHT_BLOCK, availableHeight)),
        coordinate = {
            x: mathAbs(mathMin(dLeft, 0)) - mathMax(dLeft, 0) + chartarea.left, //dLeft + chartarea.left
            y: mathAbs(mathMin(dTop, 0)) - mathMax(dTop, 0), //dTop
            w: w + PX,
            h: h - borderTop + PX,
            align: hAlign
        };

    if (hAlign === RIGHT_ALIGN) {
        coordinate.x += availableWidth - w;
    } else if (hAlign === CENTER_ALIGN) {
        coordinate.x += (availableWidth - w) / 2;
    }

    if (vAlign === BOTTOM_ALIGN) {
        coordinate.y +=
            plotarea.top + plotarea.height + (plotarea.bottom - mathMax(chartarea.bottom, 0) - h) - borderWidthTop - borderWidthRight;
    } else {
        coordinate.y += chartarea.top;
    }
    return coordinate;
};
