// @zohocharts/zc/src/layout/layout.js
import { LEFT_ALIGN, RIGHT_ALIGN, CENTER_ALIGN, TOP_ALIGN, BOTTOM_ALIGN, HORIZONTAL_ALIGN } from '../core/align-constants';
import { d3_sum, d3_max, d3_descending, d3_range } from '../core/d3-constants';
import { isSvgElement } from '../utils/zcsvgutils';
import { bbox, splat, simpleClone, isArray, invokeFunction, objectEntries, arrayFind, getPropVal, perToPx } from '../utils/zcutils';
import { isFunction } from '../utils/function-utils';
import { getLegendAlignment as getAlignment } from '../legend/legend-utils';

let WIDTH_MEASURE = 'width',
    HEIGHT_MEASURE = 'height';

export class Layout {
    constructor(units, options) {
        let { area, order, getPosition, callbackArgs = [] } = options;
        callbackArgs = splat(callbackArgs);

        this.units = units.slice(0);
        this.area = area;
        this.getPosition = getPosition;
        this.order = order;
        this.callbackArgs = callbackArgs;
        this.areaToBeUpdated = [area];
    }

    create() {
        let self = this,
            units = this.units;

        this.elementBound = {};
        this.HEIGHT = id => this.elementBound[id].height;
        this.WIDTH = id => this.elementBound[id].width;

        if (!units.length) {
            return;
        }

        this.findBound(units);
        this.prepare();
        let items = this.items;
        units = this.units;

        items.forEach((_, i) => {
            let unit = _.unit;
            if (unit.enabled && !getAlignment(getConfig(unit)).floating && !self.canFit(_)) {
                self.redraw(_, i);
                self.findBound(units[i]);
            }
        });

        this.position(units);
    }

    findBound(units) {
        let bb = this.elementBound;
        units = splat(units);

        units.forEach(_ => {
            let element = _.element.node(),
                { marginLeft, marginTop, marginRight, marginBottom } = getConfig(_),
                b = isSvgElement(element) ? bbox(element) : element.getBoundingClientRect();

            bb[_.id] = {
                width: b.width + marginLeft + marginRight,
                height: b.height + marginTop + marginBottom
            };
        });
    }

    prepare() {
        let _HEIGHT = this.HEIGHT,
            _WIDTH = this.WIDTH;

        let orderedUnits = (this.units = this.getUnitOrder());
        this.items = orderedUnits.map(function(_) {
            return { id: _.id, unit: _ };
        });

        getFormula(orderedUnits, this.items, _WIDTH, _HEIGHT, this.area);
    }

    canFit(_) {
        let callbackArgs = this.callbackArgs,
            _HEIGHT = this.HEIGHT,
            _WIDTH = this.WIDTH,
            id = _.id,
            w = _.width(id, ...callbackArgs),
            h = _.height(id, ...callbackArgs),
            newSize = [];

        if (_WIDTH(id) > w) {
            newSize[0] = w;
        }

        if (_HEIGHT(id) > h) {
            newSize[1] = h;
        }

        _.update = newSize;
        return !newSize.length;
    }

    redraw(_, i) {
        let units = this.units,
            { marginLeft, marginTop, marginRight, marginBottom } = getConfig(units[i]),
            newSize = _.update,
            w,
            h;

        if (newSize[0]) {
            w = newSize[0] - marginLeft - marginRight;
        }

        if (newSize[1]) {
            h = newSize[1] - marginTop - marginBottom;
        }

        units[i].component.resize(w, h);
    }

    position(units) {
        let self = this,
            { items, elementBound, getPosition, area: chartArea } = self,
            layoutArea = simpleClone(chartArea),
            bag = [],
            bagBound = [],
            prevSide,
            // update function
            updateUnits = side => {
                let newLayoutArea = simpleClone(layoutArea),
                    measureName = side === LEFT_ALIGN || side === RIGHT_ALIGN ? HEIGHT_MEASURE : WIDTH_MEASURE,
                    changeInAlign,
                    changeInCenterAlign,
                    //
                    blocks = getBlocks(bag, bagBound, chartArea, measureName, items),
                    set1 = [],
                    set2 = [];

                blocks.forEach(_set => {
                    changeInAlign = getChangeInAlign(side, bagBound.filter((b, i) => _set.indexOf(i) > -1), newLayoutArea);
                    let _newLayoutArea = simpleClone(newLayoutArea);

                    _set.forEach(i => {
                        let unit = bag[i],
                            config = getConfig(unit),
                            element = unit.element,
                            { hAlign, vAlign } = getAlignment(config),
                            newConfig = { ...config, ...changeInAlign };

                        if (hAlign === CENTER_ALIGN || vAlign === CENTER_ALIGN) {
                            if (!changeInCenterAlign) {
                                changeInCenterAlign = getChangeInCenterAlign(side, bagBound.slice(i, bagBound.length), _newLayoutArea);
                            }

                            if (side === LEFT_ALIGN || side === RIGHT_ALIGN) {
                                vAlign = newConfig.vAlign = changeInCenterAlign.vAlign;
                            } else {
                                hAlign = newConfig.hAlign = changeInCenterAlign.hAlign;
                            }
                        }

                        let xy = getPosition(element, newConfig, _newLayoutArea),
                            //inner block
                            alignName = side === LEFT_ALIGN || side === RIGHT_ALIGN ? vAlign : hAlign,
                            mName = side === LEFT_ALIGN || side === RIGHT_ALIGN ? HEIGHT_MEASURE : WIDTH_MEASURE;
                        _updateArea(alignName, bagBound[i][mName], _newLayoutArea);

                        Layout.positionElement(element, xy[0], xy[1]);
                        set1.push(bagBound[i]);
                    });

                    let val = getMaxOccupied(side, set1);
                    _updateArea(side, val, newLayoutArea);

                    set2.push(val);
                    set1 = [];
                });

                let val = d3_sum(set2);
                _updateArea(side, val, layoutArea);
                self.updateArea(side, val);

                bag = [];
                bagBound = [];
            };

        units.forEach(unit => {
            let { enabled, id, element } = unit,
                config = getConfig(unit);

            if (enabled) {
                let side = getSide(config);

                if (side) {
                    if (prevSide && prevSide !== side) {
                        updateUnits(prevSide);
                    }

                    let bb = elementBound[id];
                    bag.push(unit);
                    bagBound.push(bb);

                    prevSide = side;
                } else {
                    let xy = getPosition(element, config, chartArea);
                    Layout.positionElement(element, xy[0], xy[1]);
                }
            }
        });

        if (prevSide) {
            updateUnits(prevSide);
        }
    }

    updateArea(side, val) {
        _updateArea(side, val, this.areaToBeUpdated);
    }

    getUnitOrder() {
        let units = this.units,
            elementBound = this.elementBound,
            customOrder = this.order,
            callbackArgs = this.callbackArgs,
            orderedIds;

        if (customOrder) {
            orderedIds = isArray(customOrder) ? customOrder : invokeFunction(customOrder, units, elementBound, ...callbackArgs);
        } else {
            orderedIds = getDefaultOrder(units, elementBound);
        }

        return orderedIds.map(id => arrayFind(units, _ => _.id === id));
    }
}

Layout.positionElement = (ele, x, y) => {
    if (isSvgElement(ele)) {
        ele.attr('transform', `translate(${x},${y})`);
    } else {
        ele.styles({
            left: `${x}px`,
            top: `${y}px`
        });
    }
};

function _updateArea(side, val, areas) {
    let measureName = _measureName(side);
    areas = splat(areas);
    areas.forEach(_ => {
        _[side] += val;
        _[measureName] -= val;
    });
}

function getFormula(units, items, _WIDTH, _HEIGHT, area) {
    let prevSide,
        bag = [],
        findFormula = side => {
            let filteredUnits = bag.map(id => arrayFind(units, _ => _.id === id)),
                filteredItems = bag.map(id => arrayFind(items, _ => _.id === id));

            applyFormula(filteredUnits, filteredItems, _WIDTH, _HEIGHT, side, area);
            bag = [];
        };

    units.forEach(unit => {
        let { enabled, id } = unit,
            config = getConfig(unit);

        if (enabled) {
            let side = getSide(config);

            if (side) {
                if (prevSide && prevSide !== side) {
                    findFormula(prevSide);
                }

                bag.push(id);
                prevSide = side;
            }
        }
    });

    if (prevSide) {
        findFormula(prevSide);
    }
}

function applyFormula(units, items, _WIDTH, _HEIGHT, side, area) {
    let splitFunc = side => {
            return id => {
                let sizeFunc = side === HEIGHT_MEASURE ? _HEIGHT : _WIDTH,
                    cur = sizeFunc(id),
                    remainingUnits = units.map(_ => _.id).filter(_ => _ !== id),
                    remainingSize = d3_sum(remainingUnits, _ => sizeFunc(_)),
                    totalSize = area[side],
                    equal = totalSize / (remainingUnits.length + 1);

                // TODO: have to consider config.maxWigth, maxHeight
                return cur <= equal ? cur : remainingSize < equal ? totalSize - remainingSize : equal;
            };
        },
        w = id => _WIDTH(id),
        h = id => _HEIGHT(id);

    if (side === LEFT_ALIGN || side === RIGHT_ALIGN) {
        h = splitFunc(HEIGHT_MEASURE);
    } else {
        w = splitFunc(WIDTH_MEASURE);
    }

    items.forEach((_, i) => {
        let config = getConfig(units[i]),
            customWidth = getCustomSize(config, 'width', area),
            customHeight = getCustomSize(config, 'height', area);

        _.width = customWidth || w;
        _.height = customHeight || h;
        _.strict = !!(customWidth || customHeight);
    });
}

function getSide(config) {
    let { vAlign, hAlign, floating, layout } = getAlignment(config);
    let isHorizontal = layout === HORIZONTAL_ALIGN;

    if (floating) {
        return;
    }

    if (isHorizontal) {
        // ---> direction
        if (vAlign === TOP_ALIGN || vAlign === BOTTOM_ALIGN) {
            return vAlign;
        } else if (vAlign === CENTER_ALIGN) {
            return hAlign;
        }
    } else {
        // ↓↓ direction
        if (hAlign === LEFT_ALIGN || hAlign === RIGHT_ALIGN) {
            return hAlign;
        } else if (hAlign === CENTER_ALIGN) {
            return vAlign;
        }
    }
}

function getMaxOccupied(side, bagBound) {
    let measureName = _measureName(side);
    return d3_max(bagBound.map(_ => _[measureName]));
}

function _measureName(side) {
    return side === LEFT_ALIGN || side === RIGHT_ALIGN ? WIDTH_MEASURE : HEIGHT_MEASURE;
}

function getDefaultOrder(units, elementBound) {
    // put units in the respective sides
    // find the occupied measure of the each side
    // sort the sides based on the occupied measure

    let orderedIds = [],
        disabledUnits = [],
        floatingUnits = [],
        blockUnits = { left: [], right: [], top: [], bottom: [] },
        blockSizes = { left: 0, right: 0, top: 0, bottom: 0 };

    units.forEach(unit => {
        let { enabled, id } = unit,
            config = getConfig(unit);

        if (enabled) {
            let side = getSide(config);

            if (side) {
                let bb = elementBound[id],
                    measureName = side === LEFT_ALIGN || side === RIGHT_ALIGN ? HEIGHT_MEASURE : WIDTH_MEASURE,
                    occupied = bb[measureName];

                blockUnits[side].push(id);
                blockSizes[side] += occupied;
            } else {
                floatingUnits.push(id);
            }
        } else {
            disabledUnits.push(id);
        }
    });

    let temp = objectEntries(blockSizes).sort((a, b) => d3_descending(a[1], b[1]));

    temp.forEach(_ => {
        let side = _[0],
            orderedBlockUnits = blockUnits[side].sort((a, b) => {
                let aOrder = getAlignOrder(units, side, a),
                    bOrder = getAlignOrder(units, side, b);

                return d3_descending(aOrder, bOrder);
            });

        orderedIds.push(...orderedBlockUnits);
    });

    orderedIds.push(...floatingUnits);
    orderedIds.push(...disabledUnits);

    return orderedIds;
}

let alignOrder = {
    top: 2,
    bottom: 1,
    left: 2,
    right: 1,
    center: 0
};

function getAlignOrder(units, side, id) {
    let unit = arrayFind(units, _ => _.id === id),
        { hAlign, vAlign } = getAlignment(getConfig(unit)),
        alignName = side === LEFT_ALIGN || side === RIGHT_ALIGN ? vAlign : hAlign;

    return alignOrder[alignName];
}

function getChangeInAlign(side, bagBound, area) {
    if (side === RIGHT_ALIGN || side === BOTTOM_ALIGN) {
        let mName = _measureName(side),
            total = getMaxOccupied(side, bagBound),
            oppositeSide = side === RIGHT_ALIGN ? LEFT_ALIGN : TOP_ALIGN;

        _updateArea(oppositeSide, area[mName] - total, area);
        return { [side === RIGHT_ALIGN ? 'hAlign' : 'vAlign']: oppositeSide };
    } else {
        return {};
    }
}

function getChangeInCenterAlign(side, bagBound, area) {
    let alignName, alignValue, mName;

    if (side === LEFT_ALIGN || side === RIGHT_ALIGN) {
        alignName = 'vAlign';
        alignValue = TOP_ALIGN;
        mName = HEIGHT_MEASURE;
    } else {
        alignName = 'hAlign';
        alignValue = LEFT_ALIGN;
        mName = WIDTH_MEASURE;
    }

    let total = d3_sum(bagBound, _ => _[mName]);
    area[alignValue] += (area[mName] - total) / 2;
    area[mName] -= total;

    return { [alignName]: alignValue };
}

function getConfig(unit) {
    return unit.config || unit.meta.config;
}

function getBlocks(bag, bagBound, chartArea, measureName, items) {
    let blocks = [],
        blockIndex = 0,
        sum = 0;

    let strict = bag.some(b => arrayFind(items, _ => _.id === b.id).strict);

    if (!strict) {
        return [d3_range(bag.length)];
    }

    bag.forEach((unit, i) => {
        sum += bagBound[i][measureName];
        if (sum - chartArea[measureName] > 1) {
            sum = 0;
            blockIndex++;
        }

        blocks[blockIndex] = blocks[blockIndex] || [];
        blocks[blockIndex].push(i);
    });

    return blocks;
}

function getCustomSize(config, measure, area) {
    let size = config[measure],
        _size;

    if (size) {
        _size = getPropVal(window, size);
        if (!isFunction(_size)) {
            _size = () => perToPx(size, area[measure]);
        }
    }

    return _size;
}
