// @zohocharts/zc/src/axis/radial/parser/radialaxis-parser.js
import { ParserRegistry } from "../../../core/parser-registry";
import { noOperation } from "../../../core/zc-constants";
import { renderDom } from "../../../mark/dom-render";
import { executeAxisOperators } from "../../axis-operator-executor";
import { getRadialCommonOption, getRadialComponent } from "./config/radial-renderer-config";
import { ZC_RADIALAXIS_MARKS_MAPPING } from "./marks/radialaxis-mark-mapping";
import { radial_parser_helpers } from "./radialaxis-parser-helpers";

export class RadialAxisParser {
    #container; #marks; #mapping; #operators; #extender;

    constructor(dependencies) {
        const { type, container, axis, config, data, ticks, subticks, range, renderer, chart, customOperator } = dependencies;
        this.KEY = 'radial';
        this.type = type;
        this.data = data;
        this.axis = axis;
        this.config = config;
        this.ticks = ticks;
        this.subticks = subticks;
        this.range = range;
        this.renderer = renderer;
        this.chart = chart;
        this.#container = container;
        this.customOperator = customOperator;
        this.#initialize();
    }

    #initialize() {
        this.setMarkMapping();
        this.commonOption = getRadialCommonOption.call(this);
        this.component = getRadialComponent.call(this);
        this.helpers = radial_parser_helpers.call(this);
        // this.collectEncode();
        this.setMarkParser(['radial']);
    }

    container() {
        return this.#container
    }
    operators() {
        return this.#operators
    }

    marks() {
        return this.#marks;
    }

    mapping() {
        return this.#mapping;
    }
    extender() {
        return this.#extender;
    }

    #collectEncode() {
        this.#operators.forEach(operator => {
            let encode = (ParserRegistry.encodes(operator) || noOperation);
            encode.call(this);
        });
    }

    #allowedMarkOperator(operator) {
        return operator ? (this.#operators = operator, operator) : this.#operators;
    }

    setMarkMapping(mapping) {
        this.#mapping = (mapping || ZC_RADIALAXIS_MARKS_MAPPING).call(this);
    }

    setMarkParser(operator) {
        let operators = this.#allowedMarkOperator(operator),
            markMapping = this.#mapping,
            customOperator = this.customOperator;
        this.#collectEncode();
        this.#marks = executeAxisOperators.call(this, this.data, markMapping, operators, customOperator);
    }

    render(parent) {
        parent = parent || this.container();
        this.#marks.map(d => {
            renderDom(d.mark, d.holder || parent)
        })
    }

    destroy() {
        this.#container = this.#marks = null;
    }
}
ParserRegistry.parser('RadialAxis', RadialAxisParser);
