// @zohocharts/zc/src/utils/animationutils.js
//$Id$
//animationutils.js

import { mergeJSON, getValByPriority, validateNullVal, isDocumentVisible } from './zcutils';
import { mathRound, mathCos, mathMax, mathPI } from '../core/zc-constants';
import { d3_select, d3_interpolate, d3_ease } from '../core/d3-constants';
import { datacount_based_intelligence } from './intelligence-utils';
import { SCATTER_CHARTNAME } from '../core/chartname-constants';
import { POINTTWEEN_ANIMATION } from '../core/animation-constants';

export function getAnimationConf(commonPlotOptions, plotoptions, chartObj, cType) {
    var animationOptions =
            plotoptions && plotoptions.animation
                ? mergeJSON(commonPlotOptions.animation, plotoptions.animation, true)
                : commonPlotOptions.animation,
        globalAnimation = chartObj.systemConf.chart.plot.animation,
        obj = {},
        easingType =
            animationOptions.easingType ||
            globalAnimation.easingType ||
            (/(scatter|bubble$|packedbubble|heatmap)/.test(cType) ? 'cubic-in-out' : null) ||
            'linear'; //bubble$ - bubblepie chart shouldnot pass this case

    if (isDocumentVisible() || chartObj.isHiddenCoainer || chartObj.disableAnimation) {
        obj.enabled = false;
    } else {
        obj.enabled = datacount_based_intelligence(chartObj, 'animation')
            ? getValByPriority([animationOptions, globalAnimation], 'enabled')
            : false;
    }
    obj.enabled = window.rendererMode == 'canvas' ? cType == 'dial' : obj.enabled;
    obj.type = animationOptions.type;
    obj.easingType = d3_ease(easingType);
    obj.duration = !obj.enabled ? 0 : validateNullVal(animationOptions.duration, +(globalAnimation.duration || defaultDuration()));
    return obj;

    function defaultDuration() {
        return /(bar|bullet|gantt|waterfall)/.test(cType)
            ? 300
            : /(line|area|arearange)/.test(cType)
            ? 600
            : /(scatter|bubble$|funnel|pyramid|heatmap)/.test(cType)
            ? 1000 //bubble$ - bubblepie chart shouldnot pass this case
            : 500;
    }
}

export function interTween(
    baseData,
    basedataShift,
    serData,
    categories,
    animationType,
    easingType,
    transitionDuration,
    rp,
    layout,
    eleEnter,
    chartName
) {
    var isPointTween = animationType == POINTTWEEN_ANIMATION;
    var dataset = rp.dataset,
        yaxiscolumnorder = rp.yaxiscolumnorder;
    var count = isPointTween ? 0 : -1;
    var len = serData.length;
    var interTweenDuration = transitionDuration; //Negative Duration not supported in D3V4, hence removed delay 17
    var duration = mathRound(interTweenDuration / (isPointTween ? categories.length - 1 : len));
    tween();

    function tween() {
        count++;
        if (isPointTween) {
            for (var j = 0; j < len; j++) {
                baseData[j] = serData[j < count ? j : mathMax(0, count - 1)];
                basedataShift[j] = serData[j < count ? j : count];
            }
        } else {
            baseData[count - 1] = serData[count - 1];
            basedataShift[count] = serData[count];
        }
        if (count < len) {
            var del = count == 1 && isPointTween ? (categories.indexOf(dataset.getX(serData[0])) - (count - 1)) * duration : 0,
                dur =
                    (isPointTween
                        ? categories.indexOf(dataset.getX(serData[count])) - categories.indexOf(dataset.getX(serData[count - 1]))
                        : 1) * duration,
                notNull = isPointTween
                    ? dataset.getY(serData[count], yaxiscolumnorder) != null &&
                      dataset.getY(serData[mathMax(0, count - 1)], yaxiscolumnorder) != null
                    : dataset.getY(serData[count], yaxiscolumnorder) != null;
            if (notNull) {
                eleEnter
                    .attrs(layout(baseData))
                    .transition()
                    .delay(del)
                    .duration(dur)
                    .ease(easingType)
                    .attrs(layout(basedataShift))
                    .on(
                        'end',
                        chartName === SCATTER_CHARTNAME
                            ? function(d, j) {
                                  if (j == len - 1) {
                                      tween();
                                  }
                              }
                            : tween
                    );
            } else {
                tween();
            }
        }
    }
}

export function flipTween(element) {
    var rad = d3_interpolate(0, mathPI);
    element.attrTween('transform', function() {
        var transform = d3_select(this).attr('transform');
        return function(t) {
            return `${transform} matrix(${mathCos(rad(t))},0,0,1,0,0)`;
        };
    });
}

export function fadeTween(element) {
    var opacity = d3_interpolate(0, 1);
    element.styleTween('opacity', function() {
        return function(t) {
            return opacity(t);
        };
    });
}

export function scaleTween(element) {
    var scaleObj = d3_interpolate(0, 1);
    element.attrTween('transform', function() {
        var transform = d3_select(this).attr('transform') || '';
        return function(t) {
            return `${transform} scale(${scaleObj(t)})`;
        };
    });
}

export function cancelAnimation(element) {
    element.interrupt();
}
