// @zohocharts/zc/src/utils/array-utils.js
import { d3_ascending, d3_bisector, d3_descending, d3_map } from '../core/d3-constants';
import { pick } from './zcutils';

/**
 *
 * @param {Array} array
 * @param {function} callback
 * @return {Array}
 */
export function arrayFindIndex(array, callback) {
    if (Array.prototype.findIndex) {
        return array.findIndex(callback);
    } else {
        // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/findIndex
        var len = array.length >>> 0,
            i = 0;

        while (i < len) {
            var kValue = array[i];
            if (callback.call(callback, kValue, i, array)) {
                return i;
            }
            i++;
        }
        return -1;
    }
}

/*
Array list
 */

export function ArrayList() {
    this.list = d3_map();
    this.list1 = d3_map();
    this.count = 0;
}
/*
@param key - Key to be be added to map
@param value - its optional. By default automatic counter will be used
 */
ArrayList.prototype.add = function(key, value) {
    let store = pick(value, this.count);
    if (!this.list.has(key)) {
        this.list1.set(store, key);
        this.list.set(key, [store, this.count++]);
    }
};

/*
@param key - used to get the value stored in the list
 */
ArrayList.prototype.get = function(key) {
    return this.list1.get(key);
};

/*
Used to get the position of the item stored in list
@param key - item to get position
@param relative - its optional used to get the value based on the counter
 */
ArrayList.prototype.indexOf = function(key, relative) {
    var value = this.list.get(key);
    if (value) {
        return relative ? value[1] : value[0];
    }
};

ArrayList.prototype.values = function() {
    return this.list.keys();
};

export function bisectArray(value, array, comparator, end, validator, isReverse) {
    let insertIdx,
        bisector = d3_bisector(comparator),
        bisect = bisector.left(array, value, 0, end);
        
    if (validator == '>') {
        insertIdx = bisect > 0 && (isReverse ? bisect < end : true) ? (array[bisect] > value ? bisect - 1 : bisect) : bisect;
    } else if (validator == '<') {
        insertIdx = bisect > 0 ? (array[bisect] < value ? bisect - 1 : bisect) : bisect;
    } else {
        insertIdx = bisect > 0 ? (value > array[bisect - 1] ? bisect - 1 : bisect) : bisect;
    }

    return insertIdx;
}
