// @zohocharts/zc/src/utils/colorutils.js
//$Id$
//colorutils.js

import { Registry } from '../core/zc-registry';
import { DataProcessor } from '../core/data-processor';
import { PIE_CHARTNAME, FUNNEL_CHARTNAME, PYRAMID_CHARTNAME, SANKEY_CHARTNAME } from '../core/chartname-constants';
import { d3_rgb, d3_scaleLinear, d3_interpolateRgb } from '../core/d3-constants';
import {
    defined,
    isUndefined,
    pick,
    mergeJSON,
    pInt,
    validateNullVal,
    invokeFunction,
    arrayFind,
    isArray,
    arrayIncludes,
    isObjectPropDefined,
    simpleClone
} from './zcutils';
import { mathMax, mathMin, TRANSPARENT, UNDEFINED } from '../core/zc-constants';
import { effects, getGradientProp } from '../effects/zceffects';
import { colorscale_helpers } from '../axis/colorscale-helpers';
import { getChartTypeByIndex, isAreaChartType, getUniqueChartTypes, isCategoryOf } from '../core/charttype-helpers';
import { RGBA_PATTERN, charttype, legend_helpers } from '../core/main-zc-init';
import { isWithinEncode, isDataActive } from '../core/data-processor/is-data-active';
// import { getDisabledOptions } from '../legend/events/legend-event-helpers';
import { getChartSeriesColor } from '../core/seriescolor-helpers';
import { isFacet } from '../axis/hierarchy-axis-utils';
import { getImageByIndex } from '../helpers/image-pallete-helpers';
// import { getLegendConfig } from '../legend/legend-helper';
import { ALLOWED_CONFIG_OPERATORS } from '../mark/operators/config/allowed-config-operators';
import { gradient_helpers } from '../effects/gradient-helpers';

export function rgb2hex(color) {
    var rgb = color.match(RGBA_PATTERN);
    return rgb && rgb.length === 5
        ? `#${`0${parseInt(rgb[1], 10).toString(16)}`.slice(-2)}${`0${parseInt(rgb[2], 10).toString(16)}`.slice(-2)}${`0${parseInt(
              rgb[3],
              10
          ).toString(16)}`.slice(-2)}`
        : color;
}
Registry.setComponent('color.rgb2hex', rgb2hex);

export function toRGBA(color, opacity) {
    var rgba = d3_rgb(color);
    rgba.opacity = isUndefined(opacity) ? 1 : opacity;
    return rgba.toString();
}

Registry.setComponent('color.toRGBA', toRGBA);

export function rgba_canvas(color, opacity) {
    opacity = pick(opacity, 1);
    var rgb = d3_rgb(color);
    return `rgba(${rgb.r},${rgb.g},${rgb.b},${opacity * rgb.opacity})`;
}

function isRgba(color) {
    return /^rgba/.test(color);
}

//NEED TO CHANGE FUNCTION NAME
export function getColorCopy(color, k, copyType, colorMode) {
    if (color === TRANSPARENT) {
        return color;
    }

    copyType = pick(copyType, 'darker');
    var d3Color = pick(colorMode, d3_rgb)(color);
    return d3Color[copyType](k).toString();
}
Registry.setComponent('color.toneColor', getColorCopy);

/**
 * @param {Boolean} customRendererMode - Canvas mode: the highlight element which is SVG for wordcloud
 */
export function getFillColor(chartObj, seriesIndex, paneid, gradientDim, dontUsePatternPallete, customRendererMode, gradientOptions, rp) {
    var defsLocation = chartObj.defsLocation,
        cType_Obj = getChartTypeByIndex(chartObj, paneid, seriesIndex),
        chartName = cType_Obj.name,
        cType = cType_Obj.value,
        commonPlotOptions = chartObj.systemConf.chart.plot.plotoptions[chartName],
        plotoptions = (chartObj.seriesdata[seriesIndex] || chartObj.seriesdata[0]).plotoptions,
        gradientOptions =
            gradientOptions ||
            (plotoptions && plotoptions.gradients
                ? mergeJSON(commonPlotOptions.gradients, plotoptions.gradients, true)
                : commonPlotOptions.gradients),
        patternPallete = chartObj.patternPallete,
        imagePallete = chartObj.imagePallete,
        imagePalleteLength = imagePallete && imagePallete.length,
        patternPalleteLength = !dontUsePatternPallete && patternPallete && patternPallete.length,
        rendererMode = customRendererMode || chartObj.plot.renderer.mode,
        indexScale = colorscale_helpers.getIndexScale(chartObj),
        rendererIdx = rp ? rp.rendererIndex : null;

    var func = function(i, d, colorGamma, customColor, fillOpacity, customGradientDim, customPatternDim, returnSVGimage, subseriesIndex) {
        var clrValue = d && chartObj.dataset.getClr(d),
            baseClr,
            gradientsEnabled = gradientOptions && gradientOptions.type;

        gradientDim = customGradientDim || gradientDim;
        i = Math.max(i, 0) || 0; // check packed bubble and sunburst

        if (customColor) {
            //if (customColor && (!chartObj.dataObject.isAxisCategory || !rp || rp.thresholdType)) {
            baseClr = customColor;
        } else if (rp && rp.paneColor && !rp.colorOverlay) {
            baseClr = rp.paneColor;
        } else {
            baseClr = getRawColor(chartObj, i, d, seriesIndex, paneid, rendererIdx);
            i = colorscale_helpers.getItemIndex(chartObj, indexScale, d, i);
            if (indexScale && (!isWithinColorRange(clrValue, chartObj) || isUndefined(i)) && d3_rgb(baseClr).opacity === 0) {
                return baseClr;
            }
        }
        if (chartName === SANKEY_CHARTNAME && chartObj.systemConf.legend && chartObj.systemConf.legend.ranges) {
            const track = chartObj.systemConf.legend.filter.track;
            baseClr = (isWithinColorRange(d[0], chartObj) && isWithinColorRange(d[1], chartObj)) ? baseClr : 
                (track ? track.fillColor : "rgb(237, 234, 234)");
        }

        baseClr = getColorByFilter(chartObj, paneid, seriesIndex, baseClr);

        // opacity
        if (rendererMode == 'canvas' && !gradientsEnabled && !(baseClr instanceof Object)) {
            baseClr = rgba_canvas(baseClr, fillOpacity);
        }

        if (patternPalleteLength && /(url|data)/.test(baseClr)) {
            return baseClr;
        }

        if (defined(colorGamma)) {
            baseClr = getColorCopy(baseClr, colorGamma).toString();
        }
        if (imagePallete && imagePallete.length && isWithinColorRange(clrValue, chartObj)) {
            // TODO: isDataActive
            let imageParser = {
                _type: 'image',
                url: getImageByIndex(chartObj, i, paneid),
                width: (customPatternDim && customPatternDim.width) || chartObj.patternImageSizes[i % imagePalleteLength][0],
                height: (customPatternDim && customPatternDim.height) || chartObj.patternImageSizes[i % imagePalleteLength][1],
                context: (customPatternDim && customPatternDim.context) || chartObj.getPatternSvg(),
                defsPath: chartObj.defsLocation,
                rendererMode
            };
            return ALLOWED_CONFIG_OPERATORS.get(imageParser._type)(imageParser, imageParser)();
        }

        if (patternPalleteLength) {
            let { useCustomPatternDimension, customPatternColor } = chartObj.userdata.legend;
            customPatternDim =
                defined(useCustomPatternDimension) && !useCustomPatternDimension ? useCustomPatternDimension : customPatternDim;

            let patternParser = {
                _type: 'pattern',
                color: baseClr,
                pattern: patternPallete[i % patternPalleteLength],
                opacity: fillOpacity,
                size: customPatternDim,
                staticColor: customPatternColor,
                context: chartObj.getPatternSvg(),
                defsPath: chartObj.defsLocation,
                rendererMode: returnSVGimage ? 'base64' : rendererMode
            };
            return ALLOWED_CONFIG_OPERATORS.get(patternParser._type)(patternParser, patternParser)();
        }

        if (
            colorscale_helpers.isQuantileColorScale(chartObj) &&
            defined(subseriesIndex) &&
            isAreaChartType(cType) &&
            !((plotoptions && plotoptions.fillColor) || commonPlotOptions.fillColor)
        ) {
            let rendererIndex = rp ? rp.rendererIndex : seriesIndex;
            gradientOptions = getGradient4ColorBand(
                chartObj,
                chartObj.renderer[rendererIndex].seriesdataAfterUpdate[subseriesIndex],
                seriesIndex
            );
        }

    

        if (gradientDim && gradientOptions && gradientOptions.type) {
            return getGradients(gradientOptions, gradientDim, baseClr, rendererMode, fillOpacity, chartName, returnSVGimage, chartObj);
        }
      
        return baseClr;
    };

    func.gradientOptions = gradientOptions;
    return func;
}

// let _map = {
//     customColor: function(operator, applied) {
//         let { paneColor, custom } = operator;
//         return custom || paneColor;
//     },
//     image: function(operator, applied) {
//         let { paneColor, custom } = operator;
//         return custom || paneColor;
//     }
// };
//
// let defaultOperators = d3_set(['image', 'pattern', 'gradient']);
// function fillParser() {
//     let fillOperatos = {};
//     if (defaultOperators.has('customColor')) {
//         fillOperatos.customColor = {
//             custom: null, //TODO
//             paneColor: rp && rp.paneColor
//         };
//     }
//
//     if (defaultOperators.has('image')) {
//         fillOperatos.image = {
//             _type: 'image',
//             image: (d, i) => getImageByIndex(chartObj, i, paneid),
//             width: (d, i) => (customPatternDim && customPatternDim.width) || chartObj.patternImageSizes[i % imagePalleteLength][0],
//             height: (d, i) => (customPatternDim && customPatternDim.width) || chartObj.patternImageSizes[i % imagePalleteLength][0]
//         };
//     }
//
//     // if (defaultOperators.has('pattern')) {
//     //     fillOperatos.baseColor = {
//     //         color: (d, i) => getRawColor(chartObj, i, d, seriesIndex, paneid)
//     //     };
//     // }
//     //
//     // if (defaultOperators.has('gradient')) {
//     //     fillOperatos.baseColor = {
//     //         color: (d, i) => getRawColor(chartObj, i, d, seriesIndex, paneid)
//     //     };
//     // }
//
//     return fillOperatos;
// }
//
// function evalFill() {}

/**
 *
 * @param gradientOptions
 * @param dimension
 * @param baseColor
 * @param type
 * @param fillOpacity
 * @param chartName
 * @param svgImage
 * @param chartInstance
 * @return {*}
 */
export function getGradients(gradientOptions, dimension, baseColor, type, fillOpacity, chartName, svgImage, chartInstance) {
    var prop = getGradientProp(gradientOptions, dimension, baseColor, fillOpacity),
        args = { offset: chartName, chartObj: chartInstance },
        gradient;

    if (type === 'svg') {
        gradient = effects.get(effects.svg_gradients, 'svgGradientFill', prop, args);
    } else {
        gradient = effects.get(effects.canvas_gradients, 'canvasGradientFill', prop, {
            offset: chartName,
            chartObj: chartInstance,
            returnSVGimage: svgImage
        });
    }

    if (svgImage || type === 'canvas') {
        return gradient;
    }

    return `url(${chartInstance.defsLocation}#${gradient})`;
}
Registry.setComponent('color.getGradients', getGradients);

/**
 * @typedef {Object} hoverStyles
 * @param {String} hoverStyles.gradient
 * @param {Object} hoverStyles.bound
 */
/**
 * This function returns the hoverStyle `gradient` and `bound` of the series based on `hoveroptions.gradients` which acts as an util for applying gradient to pattern.
 * @param {*} data 
 * @param {*} index 
 * @param {*} chartObj 
 * @returns {hoverStyles} 
 */
export function getHoverGradients(data, index, chartObj, rendererMode = 'svg') {
    const hoverStyles = {}, 
        rendererConf = chartObj.rendererConf[index] || chartObj.rendererConf[0],
        rp = rendererConf?.params,
        hoveroptions = rp?.gvp?.('hoveroptions');
        
    let hoverGradients = hoveroptions?.gradients;
    if (hoverGradients && (hoverGradients.enabled || typeof hoverGradients !== 'object')) {
        const defaultGradients = simpleClone(rp.gvp('gradients')) , // issue when changed
            chartName = chartObj.chartTypes[index],
            seriesColor = chartObj.seriesColor[index],
            gradientDim = gradient_helpers.getDimension(chartObj, index),
            args = {
                def: { gradients: defaultGradients }, gradientDim, 
                chartName, seriesColor
            }
        
        hoverGradients = (typeof hoverGradients === 'object') ? 
            (hoverGradients.type && hoverGradients.options) ? hoverGradients : defaultGradients
            : invokeFunction(hoverGradients, data, index, chartObj, args);
        // return type url or null -> seriesColor or gradients object.
        const gradientURL = !hoverGradients ? seriesColor 
            : (typeof hoverGradients === 'string') ? hoverGradients
                : getGradients(hoverGradients, gradientDim, seriesColor, rendererMode, null, chartName, null, chartObj);

        hoverStyles.gradient = gradientURL || seriesColor;
        hoverStyles.bound = rp.bound;
    }
    return hoverStyles;
}
/*
 * @param {Object} chartObj
 * @param {Number} seriesIndex
 * @returns {Color}
 */
export function getStrokeColor(chartObj, seriesIndex, paneid, gradientDim, rp) {
    var defsLocation = chartObj.defsLocation,
        cType_Obj = getChartTypeByIndex(chartObj, paneid, seriesIndex),
        cType = cType_Obj.value,
        chartName = cType_Obj.name,
        isAreaChart = isAreaChartType(cType),
        isLineChart = cType === charttype.line,
        plotoptions = (chartObj.seriesdata[seriesIndex] || chartObj.seriesdata[0]).plotoptions,
        commonPlotOptions = chartObj.systemConf.chart.plot.plotoptions[chartName],
        gradientOptions =
            isLineChart && plotoptions && plotoptions.gradients
                ? mergeJSON(commonPlotOptions.gradients, plotoptions.gradients, true)
                : commonPlotOptions.gradients,
        hasQuantileScale = colorscale_helpers.isQuantileColorScale(chartObj),
        rendererMode = chartObj.plot.renderer.mode,
        rendererIndex = rp ? rp.rendererIndex : seriesIndex,
        clr;

    return function(i, d, customColor, colorGamma, opacity, subseriesIndex) {
        var hasColorBand =
                hasQuantileScale &&
                defined(subseriesIndex) &&
                (isLineChart || isAreaChart) &&
                !((plotoptions && plotoptions.strokeColor) || commonPlotOptions.strokeColor),
            hasGradient = isLineChart && gradientOptions.type;

        if (hasColorBand) {

            gradientOptions = getGradient4ColorBand(
                chartObj,
                chartObj.renderer[rendererIndex].seriesdataAfterUpdate[subseriesIndex],
                seriesIndex
            );
        }

        if (gradientDim && (hasColorBand || hasGradient)) {
            var baseClr = customColor || getColorByIndex(chartObj, seriesIndex, paneid),
                prop = getGradientProp(gradientOptions, gradientDim, baseClr, opacity),
                args = { offset: chartName, chartObj: chartObj },
                gradientEle =
                    rendererMode == 'svg'
                        ? effects.get(effects.svg_gradients, 'svgGradientFill', prop, args)
                        : effects.get(effects.canvas_gradients, 'canvasGradientFill', prop, { offset: chartName, chartObj: chartObj });

            if (rendererMode == 'canvas') {
                return gradientEle || baseClr;
            }

            return `url(${defsLocation}#${gradientEle})`;
        }

        if (customColor) {
            //if (customColor && (!rp || rp.thresholdType)) {
            clr = customColor;
        } else {
            i = Math.max(i, 0) || 0; // check packed bubble and sunburst
            var baseClr = getRawColor(chartObj, i, d, seriesIndex, paneid, rendererIndex);
            if (defined(colorGamma)) {
                baseClr = getColorCopy(baseClr, colorGamma).toString();
            }
            clr = baseClr;
        }

        clr = getColorByFilter(chartObj, paneid, seriesIndex, clr);

        if (rendererMode == 'canvas' && !(clr instanceof Object)) {
            clr = rgba_canvas(clr, opacity);
        }
        return clr;
    };
}

/*
 * @param {Number} clr
 * @param {Object} chartObj
 * @param {Scale} colorScale
 * @param {Array} colorScale_domain
 * @returns {Boolean}
 */
export function isWithinColorRange(clr, chartObj, colorScale, colorScale_domain, range) {
    if (!colorScale) {
        colorScale = isObjectPropDefined(chartObj, 'axes.clr.scale', true);
    }

    let config = chartObj.systemConf.legend,
        nullAsCondition = val => !(isUndefined(config.colorBand.nullAs) && val == null);

    return isWithinEncode(clr, 'clr', chartObj, colorScale, config, { condition: nullAsCondition, range });
}
Registry.setComponent('color.isWithinColorRange', isWithinColorRange);

function getDisabledColor(d, color, chartObj) {
    if (!d || (!d.dummy && d.disabled)) {
        return color;
    }

    let legendOptions = chartObj.systemConf.legend;
    let clr = chartObj.dataset.getClr(d);
    if (!isWithinColorRange(clr, chartObj)) {
        return _disabledColor(color, legendOptions.filter.track);
    }

    if (chartObj.dataObject.hasZaxis) {
        let scaleComponent = arrayFind(chartObj.scales, _ => _.id === 'z-0'),
            config = legend_helpers.getLegendConfig(scaleComponent.encode, scaleComponent.key, chartObj),
            isWithinSizeEncode = Registry.getComponent('size.isWithinSizeEncode');

        if (!isWithinSizeEncode(d, chartObj)) {
            return _disabledColor(color, Registry.getComponent('legend.event.helpers.getDisabledOptions')(config).track);
        }
    }

    return true;
}

function _disabledColor(color, config = {}) {
    var excludeOpacity = validateNullVal(config.fillOpacity, 1),
        hiddenColor = config.fillColor || TRANSPARENT;

    var rgba = d3_rgb(hiddenColor);
    excludeOpacity = rgba.opacity * excludeOpacity;

    if (excludeOpacity === 0) {
        return TRANSPARENT;
    } else {
        var interpolateClr = d3_scaleLinear()
            .domain([0, 1])
            .range([color, rgb2hex(hiddenColor)])
            .interpolate(d3_interpolateRgb);
        return interpolateClr(excludeOpacity);
    }
}

/*
 * @param {Object} chartObj
 * @param {Number} i
 * @param {Array} d
 * @param {Number} seriesIndex
 * @returns {Color} 1)Threshold 2)Color scale 3)Index based
 */
export function getRawColor(chartObj, i, d, seriesIndex, paneid, rendererIdx) {
    paneid = paneid || (d && d.paneid);
    let renderedIdx = pick(rendererIdx, seriesIndex);
    var baseClr = getColorByIndex(chartObj, i, paneid),
        chartType = getChartTypeByIndex(chartObj, paneid, 0).value,
        isAxisCategory = isCategoryOf(chartObj, chartType, 'axis'),
        rp = isAxisCategory && defined(renderedIdx) && chartObj.rendererConf && chartObj.rendererConf[renderedIdx] && chartObj.rendererConf[renderedIdx].params; // TODO: PIVOT

    let colorFinder = (chartObj.seriesdata[pick(seriesIndex, i)] || chartObj.seriesdata[0]).colorFinder,
        color;

    if (defined(colorFinder)) {
        color = invokeFunction(colorFinder, chartObj, i, d, seriesIndex);
        if (defined(color)) {
            return color;
        }
    }

    if (rp && rp.categoryColorConfig) {
        color = colorscale_helpers.getWaterfallColor(d, rp.categoryColorConfig, baseClr);
    } else if (rp && rp.colorOverlay) {
        color = getColorBythreshold(chartObj, rp, d, baseClr);
    } else if (d && colorscale_helpers.isQuantileColorScale(chartObj)) {
        color = getColorByScale(chartObj, d);
    } else {
        if (d && d.arcColor) {
            color = isArray(d.arcColor) ? d.arcColor[0] : d.arcColor;
        } else {
            color = baseClr;
        }
    }

    if (!isDataActive(d, chartObj, seriesIndex)) {
        color = getDisabledColor(d, color, chartObj);
    }

    color = getColorByFilter(chartObj, paneid, seriesIndex, color);

    return color;
}
Registry.setComponent('color.getRawColor', getRawColor);

/*
 * @param {Object} chartObj
 * @param {Number} i
 * @returns {Color} based on index
 */
export function getColorByIndex(chartObj, i, paneid, legendOrder) {
    if (chartObj.imagePallete) {
        return getImageByIndex(chartObj, i, paneid);
    }

    if (chartObj.dataObject && chartObj.dataObject.categoryHierarchy && legendOrder) {
        var legendOptions = chartObj.dataObject.hierarchical.legendOptions[legendOrder];
        var colors = chartObj.imagePallete || legendOptions.colors;
        return colors[i % colors.length];
    }
    var seriesColor = getChartSeriesColor(chartObj, paneid, i);

    return seriesColor;
}
Registry.setComponent('color.getColorByIndex', getColorByIndex);

/*
 * @param {Object} chartObj
 * @param {Object} rp
 * @param {Array} d
 * @param {Color} baseClr
 * @returns {Color} color overlay
 */
function getColorBythreshold(chartObj, rp, d, baseClr) {
    var yaxiscolumnorder = rp.yaxiscolumnorder,
        threshold = rp.threshold,
        thresholdType = rp.thresholdType,
        dataset = chartObj.dataset;

    var thresholdValue = thresholdType == 'x' ? dataset.getX(d) : dataset.getY(d, yaxiscolumnorder);
    return rp.colorOverlay(threshold, thresholdType, thresholdValue, baseClr);
}

/*
 * @param {Object} chartObj
 * @param {Array} d
 * @returns {Color} based on color scale
 */
export function getColorByScale(chartObj, d, scale, dataindex) {
    var clrValue = defined(dataindex) ? d[dataindex] : chartObj.dataset.getClr(d),
        clr = chartObj.axes.clr,
        colorScale = scale || clr && clr.scale;

    if (clrValue === UNDEFINED) {
        clrValue = null;
    }
    if (!colorScale) { return };
    let domain = colorScale.domain();
    let value = domain[0] === domain[1] ? colorScale.range()[0]: colorScale(clrValue)  //* considering normalized value as 0, if domain extremes are same
    if (!defined(clrValue) && value === invokeFunction(colorScale.unknown)) { value = colorScale(0) }  //* considering null value as 0, if not in scale
    return value;
}

/*
 * @param {Color} color
 * @returns {Color} grayscale color
 * Source: www.tannerhelland.com/3643/grayscale-image-algorithm-vb6/
 */
export function getGrayScaleColor(color) {
    var rgba = d3_rgb(color);
    var gray = pInt((rgba.r + rgba.g + rgba.b) / 3); // Method 2: Gray = (Red * 0.2126 + Green * 0.7152 + Blue * 0.0722), Method 3: (Red * 0.299 + Green * 0.587 + Blue * 0.114)
    return `rgba(${gray},${gray},${gray},${rgba.opacity})`;
}

export function isMultiColoring(commonPlotOptions, chartObj) {
    return commonPlotOptions.multiColoring && !chartObj.legendarea.show;
}

function getGradient4ColorBand(chartObj, serData, seriesIndex) {
    var rotated = chartObj.axes.rotated,
        reversed = chartObj.systemConf.chart.axes.xaxis.reversed,
        xscale = chartObj.axes.x.scale,
        ordinalRange = xscale.range(),
        hiddenColor = 'rgba(255,255,255,0)',
        isDiscrete = chartObj.systemConf.legend.colorBand.colorModeOnPlot === 'discrete',
        bandWidthBy2 = xscale.bandwidth ? xscale.bandwidth() / 2 : 0,
        sc = d3_scaleLinear()
            .domain([ordinalRange[0] + bandWidthBy2, ordinalRange[ordinalRange.length - 1] - bandWidthBy2])
            .range([0, 100]);

    var processedColors = [],
        processedOffsets = [],
        prevOffset,
        prevColor,
        prevHidden,
        curOffset,
        curColor,
        curHidden;

    serData.forEach(function(d) {
        curOffset = +sc(xscale(chartObj.dataset.getX(d)) + bandWidthBy2).toFixed(2);
        curColor = getColorByScale(chartObj, d);

        if (!isDataActive(d, chartObj, seriesIndex)) {
            curColor = getDisabledColor(d, curColor, chartObj);
        }

        curHidden = curColor === TRANSPARENT;
        if (curHidden) {
            curColor = hiddenColor;
        }

        if (!isDiscrete && (prevHidden || curHidden)) {
            if (defined(prevOffset)) {
                processedOffsets.push(`${mathMin(100, prevOffset + 0.1)}%`); // #ZC1328: Offset should be within 0 and 100 for CANVAS
                processedColors.push(hiddenColor);
            }

            processedOffsets.push(`${mathMax(0, curOffset - 0.1)}%`); // #ZC1328: Offset should be within 0 and 100 for CANVAS
            processedColors.push(hiddenColor);
        }

        // For discrete
        if (isDiscrete && prevColor) {
            if (prevHidden) {
                processedOffsets.push(`${mathMax(0, curOffset - 0.1)}%`);
                processedColors.push(prevColor);
            } else if (curHidden) {
                processedOffsets.push(`${mathMin(100, prevOffset + 0.1)}%`);
                processedColors.push(curColor);
            } else {
                var by2 = (curOffset + prevOffset) / 2;
                processedOffsets.push(`${by2}%`);
                processedColors.push(prevColor);

                processedOffsets.push(`${by2}%`);
                processedColors.push(curColor);
            }
        }

        processedOffsets.push(`${curOffset}%`);
        processedColors.push(curColor);

        prevOffset = curOffset;
        prevColor = curColor;
        prevHidden = curHidden;
    });

    var options = {
        type: 'linear',
        options: {
            linear: {
                gradientUnits: 'userSpaceOnUse',
                x1: 0,
                x2: 0,
                y1: 0,
                y2: 0,
                stopColor: processedColors,
                stopOffset: processedOffsets
            }
        }
    };

    var key = rotated ? (reversed ? 'y2' : 'y1') : reversed ? 'x1' : 'x2';
    options.options.linear[key] = 100;
    return options;
}

export function getIndexForItemColor(itemIndex, seriesIndex, extraProps, chartObj) {
    /**data - actual subseries data
     * itemIndex - renderingOrder index
     * seriesIndex - series index
     * extraProps: {
     *     multiColoring
     *     reverseColor
     *     dataUpdate
     * }
     * chartObj
     *
     */
    var data_length = extraProps.dataCount,
        actualItemIndex = extraProps.reverseColor
            ? data_length - 1 - itemIndex
            : (chartObj.isAxisCategory ||
                  chartObj.isNonAxisMultiSeriesCategory ||
                  chartObj.dataObject.isPolarAxisCategory ||
                  chartObj.isGeoMap) &&
              !extraProps.multiColoring
            ? seriesIndex
            : itemIndex;
    return actualItemIndex;
}
Registry.setComponent('color.getIndexForItem', getIndexForItemColor);

export function hasColorColumnInNonAxis(chartObj, paneid) {
    if (chartObj.isGeoMap) {
        return false;
    }

    if (!isFacet(chartObj)) {
        return false;
    }

    let chartNames = defined(paneid)
        ? chartObj.dataObject.hierarchical.gridOptions[paneid].chartTypes.map(_ => _.name)
        : getUniqueChartTypes(chartObj).values();

    return (
        chartObj.userdata.facets.color.length &&
        chartNames.some(chartName => arrayIncludes([PIE_CHARTNAME, FUNNEL_CHARTNAME, PYRAMID_CHARTNAME], chartName))
    );
}

function getColorByFilter(chartObj, paneid, seriesIndex, baseClr) {
    if (isSeriesFiltered(chartObj, paneid, seriesIndex)) {
        let legendOrder = chartObj.dataObject.hierarchical.gridOptions[paneid].legendOrder || 0,
            legendConfig = chartObj.legends.get(`clr-${legendOrder}`).meta.config;

        return _disabledColor(baseClr, legendConfig.filter.track);
    }

    return baseClr;
}

export function isSeriesFiltered(chartObj, paneid, seriesIndex) {
    let chartName = getChartTypeByIndex(chartObj, paneid, 0).name;

    if (
        isFacet(chartObj) &&
        paneid &&
        !(arrayIncludes([PIE_CHARTNAME, FUNNEL_CHARTNAME, PYRAMID_CHARTNAME], chartName) && !chartObj.userdata.facets.color.length)
    ) {
        let currentSeriesData = DataProcessor.helpers.getSeriesdata(chartObj, paneid, seriesIndex);
        return currentSeriesData.filtered;
    }
}

// function() {
//     if (imagePallete) {
//         add {}
//     }
//     if (patternPallete) {
//         add {}
//     }
//     if (gradient) {
//         add {}
//     }
//     default color {}
// }
