// @zohocharts/zc/src/utils/dashstyles.js
//$Id$
//shortdash, shortdot, shortdashdotdot, dotted, dashed, longdash, dashdot, longdashdot, longdashdotdot
//dashStyles.js

import { defined, isArray, pick } from './zcutils';
import { d3_ease_linear } from '../core/d3-constants';

export var dashStyles = {
    solid: '',
    shortdash: '1.5,1.5',
    shortdot: '0.1,1.5',
    dashed: '2.5,2.5',
    longdash: '5,5',
    dotted: '0.1,2.5',
    shortdashdotdot: '1.5,2,0.1,2,0.1,2',
    dashdot: '2.5,2.5,0.1,2.5',
    longdashdot: '5,5,0.1,5',
    longdashdotdot: '5,2.5,0.1,2.5,0.1,2.5'
};

export function getDashArray(dashStyleName, strokeWidth, scaleRatio) {
    var inList = defined(dashStyles[dashStyleName]),
        dashStyle = pick(dashStyles[dashStyleName], dashStyleName, null),
        _;

    if (!defined(dashStyle) || dashStyleName === 'solid') {
        return [];
    }

    strokeWidth = inList ? pick(strokeWidth, 1) : 1;
    if (scaleRatio) {
        strokeWidth *= scaleRatio;
    }

    if (isArray(dashStyle)) {
        _ = dashStyle;
    } else {
        _ = dashStyle.split(',');
    }
    return _.map(function(e) {
        return +e * strokeWidth;
    });
}

export function getLineCap(lineCap, dashStyle) {
    return lineCap || (dashStyle && /dot/g.test(dashStyle) ? 'round' : null);
}

export function applyMarquee(element, dur, strokeDashArray, chart) {
    var dashoffset = strokeDashArray.reduce(function(prev, current) {
        return current + prev;
    }, 0);

    element
        .attr('stroke-dasharray', strokeDashArray)
        .attr('stroke-dashoffset', dashoffset)
        .transition()
        .duration(dur)
        .ease(d3_ease_linear)
        .attr('stroke-dashoffset', 0)
        .on('end', function() {
            if (chart && chart.chartRendered) {
                applyMarquee(element, dur, strokeDashArray, chart);
            }
        });
}
