// @zohocharts/zc/src/utils/marker-parser.js
//$Id$
import { Registry } from '../core/zc-registry';
import { Mark } from '../mark/mark-utils';
import { d3_select } from '../core/d3-constants';
import { toCapitalize, splat, getDefsLocation, getVal, pick } from './zcutils';

function MarkerParser(spec) {
    let { selection, data, event } = spec;
    let mark = Mark('point', selection, data);
    mark.config = convertConfig(spec);
    mark.event = event;
    return mark;
}

function convertConfig(spec) {
    let { config, prefix, defsPath = getDefsLocation(), context = d3_select(document.body).select('svg.svgPattern') } = spec,
        names = ['fillOpacity', 'strokeOpacity', 'strokeWidth', 'dashStyle', 'lineCap'],
        colors = ['fillColor', 'strokeColor'],
        _static = ['size', 'x', 'y', 'angle', 'imageUrl', 'rotate', 'translateX', 'translateY'],
        _config = {};

    names.forEach(_ => (_config[_] = prefix ? pick(config[prefix + toCapitalize(_)], config[_]) : config[_]));
    colors.forEach(_ => (_config[_] = (prefix ? config[prefix + toCapitalize(_)] : config[_]) || config.color));
    _static.forEach(_ => (_config[_] = config[_]));
    _config.shape = config.shape || config.symbol;
    _config.dashStyle = _config.strokeWidth ? _config.dashStyle : null;

    let url = config.url || config.imageUrl;
    if (url) {
        let imageSize = splat(getVal(config.size))[0];
        _config.fillColor = {
            url,
            context,
            defsPath,
            width: imageSize,
            height: imageSize,
            x: -imageSize / 2,
            y: -imageSize / 2,
            _type: 'image'
        };
    }

    return _config;
}

Registry.setComponent('marker.parser', MarkerParser);
export { MarkerParser };
