// @zohocharts/zc/src/utils/zcsvgutils.js
//$Id$

//zcsvgutils.js

import { Registry } from '../core/zc-registry';
import { defined, each, getValByPriority, isArray, utils, toArray, bbox, decodeHTML, encodeHTML } from './zcutils';
import { mathMax, mathCeil, } from '../core/zc-constants';
import { $ZCG } from '../core/zcglobal';
import { getDefaultBoundingClientRect, getTspanBoundingClientRect } from '../axis/textBound';
import { tempCanvasContext, tempSVGPath } from '../core/main-zc-init';
import { d3_namespaces, d3_select } from '../core/d3-constants';
import { effects } from '../effects/zceffects';
import { getDashArray } from './dashstyles';
import { translateTspan } from '../label/tspan-utils';
import { copyDefaultTickAttrs } from '../axis/axisutils';
import { PolygonPathGenerator } from './geometry/polygonPathGenerator';
import { _dataEncoded, dataEncoded } from '../core/zcconf';

export function appendEle(parentNode, nodeName, data, selector, selectorAttr, selectorVal, styleObj, dataKeyFunc, options = {}) {
    let { enter_update_exit, insertBefore, remove = true } = options;
    selector = selector || '';
    selectorVal = selectorVal || '',
    selectorAttr = selectorAttr || 'zc';

    let node = parentNode.selectAll(nodeName + selector + selectorVal).data(data, dataKeyFunc),
        nodeEnter = insertBefore ? node.enter().insert(nodeName, insertBefore) : node.enter().append(nodeName),
        nodeUpdate = nodeEnter.merge(node).attr(selectorAttr, selectorVal),
        nodeExit = node.exit();

    // if (selectorAttr) {
    //     nodeUpdate.attr(selectorAttr, selectorVal);
    // }

    if (styleObj) {
        nodeUpdate.styles(styleObj);
    }

    if (remove) {
        nodeExit.remove();
    }

    return enter_update_exit ? { enter: nodeEnter, update: nodeUpdate, exit: nodeExit } : nodeUpdate;
}
Registry.setComponent('domUtils.appendElement', appendEle);

export function trimText(textObj, text, textBounds, edge, orient, pattern, marginValue, max, parent, deepTrim, marginOverflowTicklabel) {
    if (text == null || !text.trim() || !textObj) {
        //#ZC676
        return text;
    }
    var maxWidth, maxHeight;
    parent = parent || textObj;
    if (max) {
        maxWidth = max.width;
        maxHeight = max.height;
    } else {
        maxWidth = edge.right - edge.left;
        maxHeight = edge.bottom - edge.top;
    }
    // var textBounds = textObj.getBoundingClientRect()
    var overflowLength = 0;

if (orient == 'left' || orient == 'right') {
    if (textBounds.top + textBounds.height - marginValue.bottom > edge.bottom) {
        overflowLength += (textBounds.top + textBounds.height - marginValue.bottom - edge.bottom) / textBounds.height;
    }
    if (textBounds.top + marginValue.top < edge.top) {
        overflowLength += (edge.top - (textBounds.top + marginValue.top)) / textBounds.height;
    }
    maxHeight =
        textBounds.height -
        mathMax(0, textBounds.top + textBounds.height - marginValue.bottom - edge.bottom) -
        mathMax(0, edge.top - (textBounds.top + marginValue.top));
} else if (orient == 'bottom' || orient == 'top') {
    if(marginOverflowTicklabel){//to overcome misalignment of trimmed ticklabel.
        var overflow = 0;
        if (textBounds.left + textBounds.width - marginValue.right > edge.right) {
            overflow = textBounds.left + textBounds.width - marginValue.right - edge.right;
        }
        if (textBounds.left + marginValue.left < edge.left) {
            overflow = mathMax(overflow, edge.left - (textBounds.left + marginValue.left));
        }
        maxWidth = 2 * ((textBounds.width / 2) - overflow);
        overflowLength += (2 * overflow) / textBounds.width;
    }
    else{
        if (textBounds.left + textBounds.width - marginValue.right > edge.right) {
            overflowLength += (textBounds.left + textBounds.width - marginValue.right - edge.right) / textBounds.width;
        }
        if (textBounds.left + marginValue.left < edge.left) {
            overflowLength += (edge.left - (textBounds.left + marginValue.left)) / textBounds.width;
        }
        maxWidth =
            textBounds.width -
            mathMax(0, textBounds.left + textBounds.width - marginValue.right - edge.right) -
            mathMax(0, edge.left - (textBounds.left + marginValue.left));
    }
}

    overflowLength = parseFloat(overflowLength.toFixed(2));
    if (overflowLength > 0) {
        var croppedTextLength = mathCeil((1 - overflowLength) * text.length);
        textObj.textContent = text.substring(0, croppedTextLength - 1) + pattern;

        var clientRect = parent.nodeName == 'tspan' ? getTspanBoundingClientRect : getDefaultBoundingClientRect;
        var newTextBounds = clientRect(parent);
        deepTrim = textObj.nodeName == 'tspan' ? false : deepTrim; //TODO
        if (deepTrim !== false && (newTextBounds.width != textBounds.width || newTextBounds.height != textBounds.height)) {
            naiveEllipsis(textObj, parent, text, maxWidth, maxHeight, newTextBounds, pattern);
        }
    }
}

export function wordWrap(text, maxWidth, maxHeight, fontHeight) {
    var words = text.split(/\s+/).reverse(),
        word,
        testLines = [],
        wrappedLines = [],
        totalHeight = fontHeight,
        maxWidth = maxWidth; //hack for setting maxwidth
    while ((word = words.pop())) {
        testLines.push(word);
        var curWordWidth = utils.fontUtils.getTextWidth(testLines.join(' '));

        if (curWordWidth > maxWidth) {
            if (testLines.length > 1) {
                testLines.pop();
                wrappedLines.pop();
                wrappedLines.push(testLines.join(' '));
            }
            if (totalHeight > maxHeight) {
                return wrappedLines;
            }
            totalHeight += fontHeight;
            testLines = [word];
            wrappedLines.push(word);
        } else {
            wrappedLines.pop();
            wrappedLines.push(testLines.join(' '));
        }
    }
    return wrappedLines;
}

export function getCorrectedRectParams(x, y, w, h, r, yReversed, axisrotated) {
    return {
        x1: yReversed && axisrotated ? x - w : x,
        y1: yReversed && !axisrotated ? y - h : y, //x1,y1 should be at top left . x2,y2 should be at bottom right .
        x2: yReversed && axisrotated ? x : x + w,
        y2: yReversed && !axisrotated ? y : y + h
    };
}
export function roundedRect(x, y, w, h, r, yReversed, axisrotated, correctionApplied) {
    let [x1, y1, x2, y2] = [x, y, w, h];
    if (!correctionApplied) {
        ({ x1, y1, x2, y2 } = getCorrectedRectParams.apply(this, toArray(arguments)));
        r = checkRadiusWithinBoundary(r, w, h);
    }

    return `M${x1},${y2 - r.bly}L${x1},${y1 + r.tly}A${r.tlx},${r.tly} 0 0 1 ${x1 + r.tlx},${y1}L${x2 - r.trx},${y1}A${r.trx},${
        r.trY
    } 0 0 1 ${x2},${y1 + r.trY}L${x2},${y2 - r.bry}A${r.brx},${r.bry} 0 0 1 ${x2 - r.brx},${y2}L${x1 + r.blx},${y2}A${r.blx},${
        r.bly
    } 0 0 1 ${x1},${y2 - r.bly}Z`;
}
Registry.setComponent('utils.roundedRect', roundedRect);

export function checkRadiusWithinBoundary(r, width, height, isCanvasMode) {
    var radius = Object.create(r);
    var top = radius.tlx + radius.trx,
        right = radius.trY + radius.bry,
        bottom = radius.brx + radius.blx,
        left = radius.tly + radius.bly;

    var z;

    if (height < left) {
        z = left / (left - height);
        (radius.tly = radius.tly - radius.tly / z), (radius.bly = radius.bly - radius.bly / z);
    }
    if (width < top) {
        z = top / (top - width);
        (radius.tlx = radius.tlx - radius.tlx / z), (radius.trx = radius.trx - radius.trx / z);
    }
    if (height < right) {
        z = right / (right - height);
        (radius.trY = radius.trY - radius.trY / z), (radius.bry = radius.bry - radius.bry / z);
    }
    if (width < bottom) {
        z = bottom / (bottom - width);
        (radius.brx = radius.brx - radius.brx / z), (radius.blx = radius.blx - radius.blx / z);
    }

    if (isCanvasMode) {
        radius.tlx = radius.tly = Math.min(radius.tlx, radius.tly);
        radius.trx = radius.trY = Math.min(radius.trx, radius.trY);
        radius.blx = radius.bly = Math.min(radius.blx, radius.bly);
        radius.brx = radius.bry = Math.min(radius.brx, radius.bry);
    }
    return radius;
}

export function applyFont(ele, hierarchy, html) {
    var fSize = utils.getVal(getValByPriority(hierarchy, 'fontSize')),
        fFamily = getValByPriority(hierarchy, 'fontFamily'),
        fWeight = getValByPriority(hierarchy, 'fontWeight'),
        fStyle = getValByPriority(hierarchy, 'fontStyle'),
        fShadow = getValByPriority(hierarchy, 'textShadow'),
        fColor = getValByPriority(hierarchy, 'fontColor'),
        txtDec = getValByPriority(hierarchy, 'textDecoration'),
        strokeColor = getValByPriority(hierarchy, 'strokeColor'),
        fontOpacity = getValByPriority(hierarchy, 'fontOpacity');

    var styleObj = {
        'font-size': defined(fSize) ? `${fSize}px` : null,
        'font-family': fFamily,
        'font-weight': fWeight,
        'font-style': fStyle,
        'text-shadow': fShadow,
        'text-decoration': txtDec,
        'fill-opacity': fontOpacity,
        [html ? 'color' : 'fill']: fColor
    };

    if (!html && strokeColor) {
        styleObj.stroke = strokeColor;
        styleObj['stroke-width'] = getValByPriority(hierarchy, 'strokeWidth') || 3;
        styleObj['paint-order'] = 'stroke';
    }

    if (defined(ele)) {
        if (strokeColor === 'contrast') {
            styleObj.stroke = null; // phantomjs trows error to wrong color values.
        }

        ele.styles(styleObj);
    } else {
        styleObj.color = fColor;
    }

    return styleObj;
}
Registry.setComponent('domUtils.applyFont', applyFont);

export function addSVGtitleEle(textEle, data) {
    textEle.append('title').text(function() {
        return data || this.parentNode.textContent;
    });
}

export function naiveEllipsis(textNode, parent, origTextContent, maxWidth, maxHeight, textBounds, pattern) {
    /*
     * If bound is less than the max, traverse from current length to original length and update the textContent
     * Else traverse from current length to 0 and update the textContent
     */
    var index = textNode.textContent.length - 1,
        pattern_length = pattern.length;
    if (textBounds.width > maxWidth || textBounds.height > maxHeight) {
        var end = 0;
        while (index >= end) {
            changeTextContent(true);
            if (textBounds.width <= maxWidth && textBounds.height <= maxHeight) {
                return;
            }
            index--;
        }
    } else {
        var end = origTextContent.length;
        while (index < end) {
            // changeTextContent();
            if (textBounds.width >= maxWidth || textBounds.height >= maxHeight) {
                index--;
                changeTextContent(false);
                return;
            }
            index++;
        }
    }

    function changeTextContent(updateBound) {
        /*
         * This will change the textContent based on 'index' value
         * And update 'textBound' with new text bound if 'dontUpdateBound' is not true
         */
        var clientRect = textNode.nodeName == 'tspan' ? getTspanBoundingClientRect : getDefaultBoundingClientRect;
        textNode.textContent = origTextContent.substring(0, index - pattern_length) + pattern;
        if (updateBound) {
            textBounds = clientRect(parent);
        }
    }
}

export function parseSVGPath(path) {
    //#ZC1741
    var arr = path.match(/[mlhvcsqtaz][^mlhvcsqtaz]*/gi);
    return arr.map(function(d) {
        d = d.trim(); // remove white space
        var func = d.match(/[a-z]/gi).shift();
        var moves = d.match(/-?[0-9]\d*(\.\d+(e-\d+)?)?/g); // exponent is really important in arcs
        if (moves) {
            moves = moves.map(function(m) {
                return parseFloat(m);
            });
        }
        return {
            function: func,
            moves: moves
        };
    });
}

export function wrapTextContent(node, text, prop, useHTML, isParsed) {
    var isMultiLine = isArray(text) && text.length > 1,
        printProperty = useHTML ? 'html' : 'text',
        parentGroup = d3_select(node.node().parentNode);
    let text_node;

    if (isMultiLine) {
        node.text(null);
        var x = node.attr('x') || 0,
            y = node.attr('y') || 0,
            dy = node.attr('dy') ? parseFloat(node.attr('dy')) : 0.71;
        var lineNumber = 0,
            lineHeight = defined(prop) && defined(prop.lineHeight) ? prop.lineHeight : dy,
            orient = defined(prop) && prop.orient;

        lineHeight =
            orient == 'top'
                ? -(text.length - 1 + lineHeight)
                : orient == 'left' || orient == 'right'
                ? -(text.length - 2 + lineHeight) / 2
                : lineHeight; //Adjust initial lineheight for left or top aligned axis

        text.forEach(function(el, i) {
            if (useHTML) {
                insertSVGEle(parentGroup, node, el, i);
            } else {
                var tspan = node.append('tspan');
                tspan
                    .attr('x', x)
                    .attr('y', y)
                    .attr('dy', function(d, i) {
                        return `${lineNumber++ + lineHeight}em`;
                    });

                tspan[printProperty](function() {
                    return el;
                });
            }
        });
    } else {
        if (useHTML && !isParsed) {
            node.text(null);
            text_node = insertSVGEle(parentGroup, node, text);
        } else {
            // wordWrap(node,100);
            node[printProperty](text);
        }
    }
    return text_node;
}

export function insertSVGEle(parent, textNode, child, order) {
    var imageRegex = /^<image.+<\/image>/i,
        tspanRegex = /^<tspan.+<\/tspan>/i,
        groupRegex = /^<g.+<\/g>/i;

    let node = textNode.node();

    if (groupRegex.test(child)) {
        var doc = new DOMParser().parseFromString(`<svg xmlns="${d3_namespaces.svg}">${child}</svg>`, 'application/xml'),
            textElements = doc.querySelectorAll('text');
        for (let i = 0; i < textElements.length; i++) {
            let node = d3_select(textElements[i]);
            if (node.attr('position') == 'auto') {
                copyDefaultTickAttrs(textNode, node);
            }
        }

        node = parent.node().appendChild(doc.querySelector('g'));
    } else if (imageRegex.test(child)) {
        if (parent.selectAll('g.image').nodes().length) {
            var doc = new DOMParser().parseFromString(`<svg xmlns="${d3_namespaces.svg}">${child}</svg>`, 'application/xml');
            parent
                .selectAll('g.image')
                .node()
                .appendChild(doc.querySelector('image'));
        } else {
            createElement('g', '.image', [textNode.data()[0]], parent, order == 0 ? 'text' : null).html(child);
        }
    } else if (tspanRegex.test(child)) {
        var doc = SVGDOMParser(child);
        // var doc = new DOMParser().parseFromString(`<svg xmlns="${d3_namespaces.svg}">${child}</svg>`, 'application/xml');
        textNode.node().appendChild(doc.querySelector('tspan'));
    } else {
        textNode.append('tspan').text(child);
    }

    return node;
}

export function getTextContent(textNode) {
    var tspan = textNode.selectAll('tspan').nodes();
    if (tspan && tspan.length) {
        var arr = [];
        tspan.forEach(function(d) {
            arr.push(d.textContent);
        });
        return arr;
    } else {
        return textNode.node().textContent;
    }
}

export function getChildNodesContent(node) {
    let childNodes = node.node().childNodes,
        arr = [];
    if (childNodes && childNodes.length) {
        childNodes.forEach(d => arr.push(d.outerHTML));
    } else {
        arr = '';
    }
    return arr;
}
function escapeXml(text) {
    return text.replace(/[&<>'"]/g, function (char) {
        switch (char) {
            case '&': return '&amp;';
            case '<': return '&lt;';
            case '>': return '&gt;';
            case '"': return '&quot;';
            case "'": return '&apos;';
            default: return char;
        }
    });
}

/**
 *
 * @param {d3.selection} element
 * @param {String} query
 * @param {Array} data
 * @param {d3.selection} parent
 * @param {selector} insertBefore
 * @param {function} dataAccessor
 */
export function createElement(element, query, data, parent, insertBefore, dataAccessor, options = {}) {
    if (!parent) {
        parent = d3_select('body');
    }

    var selectorAttribute,
        selectorValue,
        selector,
        selectorHash = {
            '.': 'class',
            '#': 'id'
        };

    //* selector can be written without ".|#" in the start(only string can be present). Default selector is taken as "."
    query = /([.|#]?)(.*)/.exec(query);
    selector = query[2] ? query[1] || '.' : undefined;
    selectorValue = query[2];
    selectorAttribute = selectorHash[selector];
    // if (/^[.#]/.test(query)) {
    //     selector = query[0];
    //     selectorValue = query.slice(1);
    //     selectorAttribute = selectorHash[selector];
    // }

    // #ZC1527
    // insertBefore - code has been moved to appendEle api

    return appendEle(parent, element, data, selector, selectorAttribute, selectorValue, null, dataAccessor, { ...options, insertBefore });
}

export function isSvgElement(element) {
    if (element.isD3Selection) {
        element = element.node();
    }

    return !!element.ownerSVGElement;
}

export function getEmptyBBox() {
    var t = {};
    t.x = t.y = t.width = t.height = 0;
    return t;
}

var fontUtils = (utils.fontUtils = {
    /**
     *
     * @param text
     * @param fontSize
     * @param fontFamily
     * @param fontStyle
     * @param fontWeight
     * @return {number}
     */
     getTextWidth: function(text, fontSize, fontFamily, fontStyle, fontWeight) {
        fontStyle = fontStyle || 'normal';
        fontWeight = fontWeight || 'normal';

        let context = tempCanvasContext()
        context.save();
        context.font = `${fontStyle} ${fontWeight} ${fontSize || 12}px ${fontFamily || $ZCG.fontFamily}`;
        var t = context.measureText(text).width;
        context.restore();
        return t;
    },

    /**
     *
     * @param fontSize
     * @param dontUseEmpirical
     * @return {*}
     */
    getTextHeight: function(text, fontSize, fontFamily, fontStyle, fontWeight, dontUseEmpirical) {
        // Empirical values found by comparing font size and bounding box height.
        // Applies to the default font family. http://jsfiddle.net/highcharts/7xvn7/
        if (dontUseEmpirical) {
            return fontSize;
        }
        return fontSize < 24 ? fontSize + 3 : ~~(fontSize * 1.195);
    }
});

utils.createElement = createElement;

/**
 *
 * @param element
 * @param styles
 * @param customStyleMap
 */
export function setStyles(element, styles, customStyleMap = {}, customAttrMap = {}) {
    // common maps

    let selectionAttrMap = {
        transform: 'transform',
        filter: 'filter',
        dashStyle: 'stroke-dasharray',
        ...customAttrMap
    };

    let selectionStyleMap = {
        fillColor: 'fill',
        innerFillColor: 'fill',
        outerFillColor: 'fill',
        fillOpacity: 'fill-opacity',
        innerFillOpacity: 'fill-opacity',
        outerFillOpacity: 'fill-opacity',
        strokeColor: 'stroke',
        innerStrokeColor: 'stroke',
        outerStrokeColor: 'stroke',
        strokeOpacity: 'stroke-opacity',
        innerStrokeOpacity: 'stroke-opacity',
        outerStrokeOpacity: 'stroke-opacity',
        strokeWidth: 'stroke-width',
        innerStrokeWidth: 'stroke-width',
        outerStrokeWidth: 'stroke-width',
        fontWeight: 'font-weight',
        fontStyle: 'font-style',
        textShadow: 'text-shadow',
        ...customStyleMap
    };

    // apply style
    each(styles, function(value, prop) {
        if (selectionAttrMap[prop]) {
            // #ZC2125
            if (prop === 'dashStyle') {
                value = getDashArray(value, styles.strokeWidth);
            }
            element.attr(selectionAttrMap[prop], value);
        } else {
            element.style(selectionStyleMap[prop] || prop, value);
        }
    });
    return element;
}

utils.setStyles = setStyles;

/**
 *
 * @param shadow
 * @param chart
 * @return {string}
 */
export function svgBoxShadow(shadow, chart) {
    if (shadow) {
        var parsedShadowOptions = shadow.split(' '),
            shadowOptions = {
                dx: parseFloat(parsedShadowOptions[0]),
                dy: parseFloat(parsedShadowOptions[1]),
                color: parsedShadowOptions[3],
                blur: parseFloat(parsedShadowOptions[2]),
                x: '-50%',
                y: '-50%',
                width: '200%',
                height: '200%'
            };

        var id = effects.get(effects.svg_filters, 'dropShadow', shadowOptions, { chartObj: chart });

        return `url(${chart.defsLocation}#${id})`;
    }
}

utils.svgBoxShadow = utils.setStyles;

export function elementToPolygonPoints(ele) {
    if (ele.isD3Selection) {
        ele = ele.node();
    }

    let len = ele.getTotalLength(),
        points = [];

    for (let k = 0; k < len; k++) {
        let p = ele.getPointAtLength(k);
        points.push([p.x, p.y]);
    }

    return points;
}

/**
 * 
 *! This function is deprecated from direct usage M2.35 due to performance issue with getPointAtLength browser API. Try `pathToPolygon` instead.
 * @param path
 * @param sample
 * @param offset
 * @return {[]}
 */
export function pathToPolygonNative(path, sample, offset = [0, 0]) {
    const pathNode = tempSVGPath();
    pathNode.setAttributeNS(null , 'd', path);
    const pathLength = pathNode.getTotalLength();

    const polygon = [],
        step = 1 / sample,
        max = 1 + step;

    for (let t = step, point; t <= max; t += step) {
        point = pathNode.getPointAtLength(t * pathLength);
        polygon.push([point.x + offset[0], point.y + offset[1]]);
    }
    return polygon;
}

/**
 *
 * This API replicates getPointAtLength property to get polygon points for data finder. 
 ** Note: Some path commands are currently supported, other path commands has to be supported for complete usage. Also need to work for relative coordinates.
 * @param {string} path - d attribute of the path element
 * @param {number} sample - number of sample from start to end point
 * @param {Number[2]} offset - points offset
 */
export function pathToPolygon(path, sample, offset = [0, 0]) {
    const polygon = [];
    let commands = parseSVGPath(path);

    let curvesCommand = ['C', 'A'],
        curvesCount = 0,
        linesCount = 0;
    commands.forEach((d) => curvesCommand.includes(d.function) ? curvesCount++ : linesCount++);
    sample -= linesCount;

    for (let i = 0, position, val; i < commands.length; i++) {
        const { function: command, moves: values } = commands[i];
        switch (command) {
            case 'M': 
            case 'L':
                polygon.push(position = values);
                break;
            case 'H':
                polygon.push(position = [values[0], position[1]]);
                break;
            case 'V':
                polygon.push(position = [position[0], values[0]]);
                break;
            case 'C':
                val = PolygonPathGenerator.cubic_bezier([...position, ...values], sample / curvesCount);
                polygon.push(...val);
                position = val[val.length - 1];
                break;
            case 'A': 
                val = PolygonPathGenerator.arc([...position, ...values], sample / curvesCount);
                polygon.push(...val);
                position = val[val.length - 1];
                break;
            case 'Z':
                polygon.push(commands[0].moves);
                break;
            default:
                val = pathToPolygonNative(`M${position.join(',')} ${command + values.join(',')}`, sample / curvesCount, offset);
                polygon.push(...val);
                position = val[val.length - 1];
        }
    }
    return polygon;
}

export function SVGDOMParser(input) {
    input = decodeNonXMLEntities(String(input));
    let domAsString = `<zc-temp>${input}</zc-temp>`,
        sandBox = new DOMParser().parseFromString(`<svg xmlns="${d3_namespaces.svg}">${domAsString}</svg>`, 'application/xml');
    return sandBox;
}
export function decodeNonXMLEntities(str) {
    let parser = new DOMParser(),
        actualDataEncodeVal = dataEncoded();
    dataEncoded(true);
    var result = str.replace(/&[#\w]+;/g, (match) => {
        let xmlDoc = parser.parseFromString(`<root>${match}</root>`, 'application/xml');
        if (!xmlDoc.querySelector('parsererror')) {
            return match;
        }
        let htmlDoc = decodeHTML(match);
        return htmlDoc;
    });
    dataEncoded(actualDataEncodeVal);
    return result
}
export function parseSVGElementsToArray(element, tag) {
    var selectedElements = toArray(element.querySelectorAll(tag));

    return selectedElements;
    // parseSVGElementsToArray(element, tag, styleProps, attrProps, isTextNode)
    //     var styleValues,
    //     attrValues,
    //     textContents = new Array(0);
    //
    // if (selectedElements.length) {
    //     selectedElements.forEach(function(t) {
    //         if (isTextNode) {
    //             textContents.push(t.textContent.trim());
    //         }
    //
    //         if (styleProps) {
    //             let style = t.style || {};
    //             styleValues = iterateObject(style, styleProps);
    //         }
    //         if (attrProps) {
    //             let style = t.style || {};
    //             attrValues = iterateObject(style, attrProps);
    //         }
    //     });
    // } else if (isTextNode) {
    //     textContents.push(element.textContent);
    // }
    // return {
    //     nodes: selectedElements,
    //     styles: styleValues,
    //     attrs: attrValues,
    //     textNodes: textContents
    // };
}

export function getTextBoundByDOM(node, content, contentType, isMultiLine) {
    var contentType = contentType || 'text',
        isMultiLine = contentType === 'html' && isMultiLine;

    node[contentType](content);
    if (isMultiLine) {
        translateTspan(node, 0);
    }

    var bound = bbox(node.node());
    return bound;
}
