// @zohocharts/zc/src/utils/zcutils.js
//$Id$
//zcutils.js
// import {mathMin} from '../core/zc-constants';
import { Registry } from '../core/zc-registry';
import * as consts from '../core/zc-constants';
import { DEVICE_PIXEL_RATIO, mathAbs } from '../core/zc-constants';
import { d3_hsl, d3_interpolateRgb, d3_rgb, d3_scaleLinear, d3_select, d3_selectAll } from '../core/d3-constants';
import { _dataEncoded } from '../core/zcconf';
import { Set } from '../core/polyfill';
import { isExpressionType, getExpressionValue } from '../mark/parser/parser-helpers';

export var utils = function() {};

/*
 * element - to which the bound box is drawn
 * component1 - node from which we start to draw the box
 * component2 - where to end the box
 * properties - outer box configuration
 */

export function drawBorderBox(element, properties) {
    element
        // .attr('x', properties.left)
        // .attr('y', properties.top)
        .attr('width', properties.width)
        .attr('height', properties.height)
        .style('fill-opacity', 0)
        .style('stroke', consts.TRANSPARENT);
}
utils.drawBorderBox = drawBorderBox;

utils.getActualTextBounds = function(fontSize, width, rotation) {
    var width = width,
        height = fontSize,
        radian = rotation * ((consts.mathPI * 2) / 360);
    var bbox = {};
    bbox.width = consts.mathAbs(height * consts.mathSin(radian)) + consts.mathAbs(width * consts.mathCos(radian));
    bbox.height = consts.mathAbs(height * consts.mathCos(radian)) + consts.mathAbs(width * consts.mathSin(radian));
    return bbox;
};
export var mergeJSON = (utils.mergeJSON = function(source1, source2, mergeConflicts, mergeIntoExists, mergeMismatchedType) {
    /*
     * Properties from the Souce2 object will be copied to Source1 Object.
     * Note: mergeIntoExists : true -- Here source 1 and source2 will get merged into source1.
     * mergeIntoExists : false -- This method will return a new merged object, Source1 and Source2 original values will not be replaced.
     * */
    source1 = source1 || {};
    source2 = source2 || {};

    var mergedJSON = mergeIntoExists != null && mergeIntoExists == true ? source1 : simpleClone(source1); // Cloning the source1
    for (var attrname in source2) {
        if (source2.hasOwnProperty(attrname)) {
            if (mergedJSON[attrname] != undefined) {
                if (source2[attrname] != null && (isProperObject(source2[attrname]) || isArray(source2[attrname]))) {
                    if (isArray(source2[attrname]) && !isArray(mergedJSON[attrname])) {
                        //check for source2 is array, but mergedJson is of string -> no need to recursive call, simply Merge the property from source 2 to source1
                        if (mergeConflicts) {
                            mergedJSON[attrname] = source2[attrname];
                        }
                    }else if (isProperObject(source2[attrname]) && (mergedJSON[attrname].constructor == String || mergedJSON[attrname].constructor == Number || mergedJSON[attrname].constructor == Boolean)) {
                        //check for source2 is object, but mergedJson is of string -> no need to recursive call, simply Merge the property from source 2 to source1
                        if (mergeConflicts) {
                            mergedJSON[attrname] = source2[attrname];
                        }
                    } else {
                        /*
                         * Recursive call if the property is an object,
                         * Iterate the object and set all properties of the inner object.
                         */
                        mergedJSON[attrname] = mergeJSON(mergedJSON[attrname], source2[attrname], mergeConflicts, mergeIntoExists);
                    }
                } else if (mergeConflicts == true && typeof mergedJSON === 'object') {
                    //Merge the property from source 2 to source1
                    mergedJSON[attrname] = source2[attrname];
                }
            } else {
                //else Add the property from source 2 to source1
                mergedJSON[attrname] = source2[attrname];
            }

            if (mergedJSON[attrname] && mergedJSON[attrname]._zc) {
                // Removing internal properties - has circular reference ZC3271
                mergedJSON[attrname] = undefined;
            }
        }
    }

    return mergedJSON;
});
Registry.setComponent('utils.mergeJSON', mergeJSON);

utils.getValueOfPlotOptions = function(obj1, obj2) {
    /*
     * obj1 -->high priority
     * obj2 -->Low priority
     */
    return function(key) {
        if (obj1 != null && obj1[key] != null) {
            return obj1[key];
        } else if (obj2 != null && obj2[key] != null) {
            return obj2[key];
        } else {
            return null;
        }
    };
};

export function getVal(value, args, self = null) {
    /*
     * If the value is function, execute it and return the value
     */
    return typeof value === 'function' ? value.apply(self, args) : value;
}
utils.getVal = getVal;

export function getParserVal(val, args, config, viewflow){
    if(typeof val === 'function'){
        return val.apply(self, args)
    }else if(isExpressionType(val)){
        return getExpressionValue(null, args || [null, null], config, null, viewflow)
    }else{
        return val;
    }
}
export function generateIdByPropValue(properties, name) {
    if (properties == null || properties.constructor != Object) {
        if (name != null) {
            return name;
        }
        return;
    }
    var id = '',
        val;
    for (var prop in properties) {
        if (properties.hasOwnProperty(prop)) {
            val = properties[prop];
            if (val && val.constructor === Object) {
                //#ZC1142 on redraw
                val = jsonStringify(val);
            }

            id += val == null || val == '' ? '$$' : `${val}_`;
        }
    }
    id = name + id;
    id = id.replace(/[.-]/g, '_');
    id = id.replace(/[`~!@#$%^&*( )|+\-=?;:'",.<>{}[\]\\/\s]/gi, '_');
    return id;
}
utils.generateIdByPropValue = generateIdByPropValue;

export function getMonochromeColor(baseClr, domain, range) {
    var hsl = d3_hsl(baseClr);

    var interpolateClr = d3_scaleLinear()
        .domain([-1, 0, 1])
        .range([d3_hsl(hsl.h, hsl.s, 0.1), hsl, d3_hsl(hsl.h, hsl.s, 0.9)])
        .interpolate(d3_interpolateRgb);
    var clrScale = d3_scaleLinear()
        .domain(domain)
        .range(range);
    return function(clr) {
        return interpolateClr(clrScale(clr));
    };
}
utils.getMonochromeColor = getMonochromeColor;

export function bbox(ele) {
    //if element is removed from DOM, firefox will throw NS_ERROR_FAILURE on getBBox(). This function returns getBBox() of the given element if that element is present in DOM, else returns {x:0, y:0, height:0, width:0}
    if (ele) {
        if (ele.isD3Selection) {
            ele = ele.node();
        }

        if (isElementVisible(ele)) {
            return ele.getBBox();
        }
    }
    return {
        x: 0,
        y: 0,
        height: 0,
        width: 0
    };
}
Registry.setComponent('domUtils.getBBox', bbox);

export var simpleClone = (utils.simpleClone = function(o) {
    var copy = o,
        k;

    if (o && typeof o === 'object') {
        copy = isArray(o) ? [] : {};
        for (k in o) {
            // if(o.hasOwnProperty(k)){
            copy[k] = simpleClone(o[k]);
            // }
        }
    }

    return copy;
});
Registry.setComponent('utils.simpleClone', simpleClone);

export function decodeHTML(text) {
    if (!_dataEncoded || text == null) {
        return text;
    }
    var tempDiv = document.createElement('div');
    tempDiv.innerHTML = encodeURI(text);
    var decodedText = tempDiv.textContent;
    return decodeURI(decodedText);
}
Registry.setComponent('domUtils.decodeHTML', decodeHTML);

export function encodeHTML(text) {
    if (_dataEncoded || text == null) {
        return text;
    }
    var tempDiv = document.createElement('div');
    tempDiv.textContent = decodeHTML(text);
    var encodedText = tempDiv.innerHTML;
    return encodedText;
}
utils.encodeHTML = encodeHTML;
Registry.setComponent('domUtils.encodeHTML', encodeHTML);

export function validateURL(url) {
    var exp_http = /https?:\/\/(www\.)?[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)/g,
        exp_nonHttp = /[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)?/gi;
    var regExp_http = new RegExp(exp_http),
        regExp_nonHttp = new RegExp(exp_nonHttp);
    return url.match(regExp_http) || url.match(regExp_nonHttp);
}
utils.validateURL = validateURL;

export function d3_values(map) {
    return map.slice(0);
}
utils.d3_values = d3_values;

export function getDefsLocation() {
    return d3_selectAll('base').node() ? location.href.split(location.search || location.hash || /[?#]/)[0] : '';
}
Registry.setComponent('domUtils.getDefsLocation', getDefsLocation);

export function isObjectPropDefined(obj, key, self) {
    var keySet = key.replace(/\[(\w+)\]/g, '.$1'); // convert indexes to properties
    keySet = keySet.replace(/^\./, ''); // strip a leading dot
    var keys = keySet.split('.');

    var childObj = keys.reduce(function(o, x) {
        return !defined(o) ? o : o[x];
    }, obj);
    return self ? childObj : defined(childObj);

    // return keys.every(function (x) {
    // 	if (typeof obj != "object" || obj === null || !x in obj)
    // 		return false;
    // 	obj = obj[x];
    // 	return true;
    // });
}
utils.isObjectPropDefined = isObjectPropDefined;

/**
 *
 * @param {Object|*} variable
 * @return {boolean}
 */
export function isUndefined(variable) {
    return !defined(variable);
}
Registry.setComponent('utils.isUndefined', isUndefined);

export function toArray(objectLikeArray) {
    return Array.prototype.slice.call(objectLikeArray);
}
Registry.setComponent('utils.toArray', toArray);

export function objectToArray(data, rule, slice) {
    var array = [];
    for (let key in data) {
        if (data.hasOwnProperty(key) && isObject(data[key])) {
            if (!rule) {
                array.push(data[key]);
            } else if (rule(data[key])) {
                array.push(data[key]);
            }

            if (slice) {
                array = array.slice(slice[0], slice[1]);
            }
        }
    }
    return array;
}
utils.objectToArray = objectToArray;

/**
 * It invokes a function with the given arguments.
 * If the first argument is true, then corresponding function will have
 * access to 'this' keyword with the value arguments[2]
 */

export function invokeFunction() {
    var args = toArray(arguments),
        advanced = args[0] === true && args.shift(),
        callback = args.shift();

    if (typeof callback === 'string') {
        var temp = getPropVal(window, callback);
        // callback = defined(temp) ? temp : callback; // for both global variable and global function support.
        callback = typeof temp === 'function' ? temp : callback; // for only global function support.
    }
    return typeof callback === 'function'
        ? advanced
            ? callback.apply(args[0], args.splice(1))
            : callback.apply(undefined, args)
        : callback;
}
Registry.setComponent('utils.invokeFunction', invokeFunction);

export function getValue(key, args) {
    if (defined(key)) {
        if (typeof key == 'function') {
            return key.call(args);
        } else {
            return key;
        }
    }
    return null;
}
utils.getValue = getValue;

export function getValByPriority(array, attr) {
    var i = 0,
        len = array.length,
        val;
    for (; i < len && (val == null || val === ''); i++) {
        if (array[i]) {
            val = array[i][attr];
        }
    }
    return val;
}
utils.getValByPriority = getValByPriority;

export function validateNullVal(actualVal, defaultVal) {
    var val = actualVal == null || actualVal === '' ? defaultVal : actualVal;

    var parsedVal = JSON.parse(val);
    val = typeof parsedVal == 'number' ? parseFloat(parsedVal) : typeof parsedVal == 'boolean' ? JSON.parse(parsedVal) : parsedVal;
    return val;
}
Registry.setComponent('utils.validateNullValue', validateNullVal);

export function validateValWithinLimit(data, prop, minPercent, maxPercent, width, height) {
    var val;
    var refLevel = consts.mathMin(width, height);
    var maxVal = (maxPercent * refLevel) / 100,
        minVal = (minPercent * refLevel) / 100;
    val = data[prop];
    val = val == null || val === '' ? null : val < minVal ? minVal : val > maxVal ? maxVal : parseFloat(val);
    return val;
}
utils.validateValWithinLimit = validateValWithinLimit;

export function isWithinRange(value, range) {
    var minVal = range[0],
        maxVal = range[1];
    return value < minVal || value > maxVal ? false : true;
}
utils.isWithinRange = isWithinRange;

/**
 * Returns true if the object is not null or undefined.
 * @param {Object} obj
 */

export function defined(obj) {
    return obj !== undefined && obj !== null;
}
Registry.setComponent('utils.defined', defined);

/**
 * Returns true if the number is positive or not.
 * @param {Object} obj
 */

export function isPositive(number) {
    return number > 0;
}
utils.isPositive = isPositive;

/**
 * Returns true if the value is null object or null string
 * @param {Object} obj
 */

export function isNull(value) {
    return value == null || value == 'null';
}
utils.isNull = isNull;

/**
 *
 * @param e
 * @returns {boolean}
 */

export function isTrue(e) {
    return e + ''.toLocaleLowerCase() === 'true';
}
utils.isTrue = isTrue;

export function shorthand_px(val) {
    /*
     * Input : 10 5 5 10
     * Output : 10px 5px 5px 10px
     */
    val += '';
    return `${val.replace(/ /g, 'px ')}px`;
}
utils.shorthand_px = shorthand_px;

/**
 * Shortcut for parseInt
 * @param {Object} s
 * @param {Number} mag Magnitude
 */
export function pInt(s, mag) {
    return parseInt(s, mag || 10);
}
utils.pInt = pInt;

/**
 * Check for string
 * @param {Object} s
 */
export function isString(s) {
    return typeof s === 'string';
}
utils.isString = isString;

/**
 * Check for object
 * @param {Object} obj
 */
export function isObject(obj) {
    return obj && typeof obj === 'object';
}
utils.isObject = isObject;

/**
 * Check for object properly, doesn't consider array instance as object
 * @param {Object} obj
 */
export function isProperObject(obj) {
    return obj && (obj.constructor == Object || obj.constructor.name == 'Object');
}
utils.isProperObject = isProperObject;

/**
 * Check for array
 * @param {any} obj
 * @param {Number} dimension - checks the dimension of the array. may be 2D or 3D or more...
 */
export function isArray(obj, dimension) {
    if (Array.isArray) {
        return Array.isArray(obj) && (!(dimension > 1) || isArray(obj[0], dimension - 1));
    }
    return !!obj && obj.constructor === Array && (!(dimension > 1) || isArray(obj[0], dimension - 1));
    // return Object.prototype.toString.call(obj) === '[object Array]';
}
Registry.setComponent('utils.isArray', isArray);

/**
 * Check for number
 * @param {Object} n
 */
export function isNumber(n) {
    return typeof n === 'number' && !isNaN(n);
}
utils.isNumber = isNumber;

export function jsonStringify(obj, replacer) {
    if (defined(obj)) {
        return JSON.stringify(obj, replacer);
    }
    return consts.UNDEFINED;
}
utils.jsonStringify = jsonStringify;

/**
 *
 * @param {Object} obj
 * @param {String} key
 * @returns {*}
 */
export var getPropVal = (utils.getPropVal = function(obj, key) {
    return isObjectPropDefined(obj, key, true);
});
/**
 *
 * @type {function(*=, *, *): *}
 */
export var setPropVal = (utils.setPropVal = function(object, keySet, value, updateCallback) {
    keySet = keySet.replace(/\[(\w+)]/g, '.$1'); // convert indexes to properties
    keySet = keySet.replace(/^\./, ''); // strip a leading dot
    var keys = keySet.split('.');
    var add = function(obj) {
        var key = keys.shift();
        if (key === null || keys.length === 0) {
            updateCallback ? updateCallback(obj, key, value) : (obj[key] = value);
            return false;
        } else if (obj[key] === null || typeof obj[key] !== 'object') {
            obj[key] = !isNaN(parseInt(keys[0])) ? [] : {};
        }
        add(obj[key]);
    };
    add(object);
    return object;
});

/**
 * changes the read-only property from an object
 * @param {any} obj - obj
 * @param {string} key 
 * @param {any} val 
 * @returns {any} - first argument(obj)
 */
export function setFrozenVal(obj, key, val) {
    delete obj[key];
    obj[key] = val;
    return obj
}

export function intersectObject(obj1, obj2) {
    if (!obj1 || !obj2) {
        return null;
    }
    var output = {},
        intersected;
    for (var prop in obj1) {
        if (obj1.hasOwnProperty(prop) && obj2.hasOwnProperty(prop)) {
            var val1 = obj1[prop],
                val2 = obj2[prop];
            if (val1 && (val1.constructor == Object || val1.constructor == Array)) {
                var i = intersectObject(val1, val2);
                if (i != null) {
                    output[prop] = i;
                    intersected = true;
                }
            } else if (val1 == val2) {
                output[prop] = val1;
                intersected = true;
            }
        }
    }
    return intersected ? output : null;
}
utils.intersectObject = intersectObject;

/**
 *
 * @param {String} color
 * @returns {boolean}
 */

function isDarkColor(color) {
    //TODO: consider alpha

    // http://stackoverflow.com/questions/2509443/check-if-uicolor-is-dark-or-bright/2509596#2509596
    // https://stackoverflow.com/questions/12043187/how-to-check-if-hex-color-is-too-black
    // https://stackoverflow.com/questions/24260853/check-if-color-is-dark-or-light-in-android
    // https://stackoverflow.com/questions/635022/calculating-contrasting-colours-in-javascript

    var rgb = d3_rgb(color);

    // return 1 - (0.299 * rgb.r + 0.587 * rgb.g + 0.114 * rgb.b) / 255 > 0.45;

    // return 0.2126 * rgb.r + 0.7152 * rgb.g + 0.0722 * rgb.r < 165; // zoho-reports version

    return (rgb.r * 299 + rgb.g * 587 + rgb.b * 114) / 1000 < 128; // tiny color
}
utils.isDarkColor = isDarkColor;

/**
 *
 * @param {String} color
 * @returns {boolean}
 */

function isLightColor(color) {
    return !isDarkColor(color);
}
utils.isLightColor = isLightColor;

/**
 *
 * @param {String} color
 * @returns {string}
 */

export function getContrastColor(color) {
    return isDarkColor(color) ? '#f8f8f8' : '#444444';
}
utils.getContrastColor = getContrastColor;

/**
 *
 * @param {*} obj
 * @returns {boolean}
 */

function isWindow(obj) {
    return obj != null && obj === obj.window;
}
utils.isWindow = isWindow;

/**
 *
 * @param elem
 * @returns {boolean|DocumentView}
 */

function getWindow(elem) {
    return isWindow(elem) ? elem : elem.nodeType === 9 && elem.defaultView;
}
utils.getWindow = getWindow;

/**
 *
 * @param {Element} elem
 * @returns {{top: number, left: number}}
 */

function offset(elem) {
    var box = elem.getBoundingClientRect(),
        doc = elem && elem.ownerDocument,
        docElem = doc.documentElement,
        win = getWindow(doc);
    return {
        top: box.top + win.pageYOffset - docElem.clientTop,
        left: box.left + win.pageXOffset - docElem.clientLeft
    };
}
utils.offset = offset;

/**
 *
 * @param {String} str
 * @returns {Array}
 */
export function stringifiedArrayReverse(str) {
    if (!str) {
        return str;
    }

    return str
        .split(',')
        .reverse()
        .join();
}
utils.stringifiedArrayReverse = stringifiedArrayReverse;

/**
 *
 * @returns {Boolean}
 */
export function isDocumentVisible() {
    return defined(document.visibilityState) && document.visibilityState !== 'visible'; //window blur state
}
utils.isDocumentVisible = isDocumentVisible;

/**
 *
 * @param {Element} element
 * @param {Element} container
 * @param {Boolean|null} partial
 * @returns {boolean}
 */

function elementIsVisible(element, container, partial) {
    var contHeight = container.offsetHeight,
        elemTop = offset(element).top - offset(container).top,
        elemBottom = elemTop + element.offsetHeight;
    return (
        (elemTop >= 0 && elemBottom <= contHeight) ||
        (partial && ((elemTop < 0 && elemBottom > 0) || (elemTop > 0 && elemTop <= contHeight)))
    );
}

utils.elementIsVisible = elementIsVisible;

/**
 * Return the first value that is defined.
 */

export var pick = function() {
    var args = arguments,
        i,
        arg,
        length = args.length;
    for (i = 0; i < length; i++) {
        arg = args[i];
        if (defined(arg)) {
            return arg;
        }
    }
    return args[length - 1];
};
utils.pick = pick;

/**
 *
 * joins one or multiple arrays.
 * if the first argument is true, then remaining arrays will copied to the first array
 * else it copies all arrays to a new array.
 * @return {Array}
 */

function arrayJoin() {
    // performance https://jsperf.com/array-prototype-push-apply-vs-concat/5
    // https://jsperf.com/arrayconcatvsarraypushapply
    // return target.concat(newObject)

    var args = toArray(arguments),
        copyToExisting = args[0] === true && args.shift(),
        target = copyToExisting ? args.shift() : [];

    if (copyToExisting && !args.length) {
        return target;
    }

    args.forEach(function(newObject) {
        var i = 0,
            c = newObject.length;
        for (; i < c; ++i) {
            target.push(newObject[i]);
        }
    });
    return target;
}
utils.arrayJoin = arrayJoin;

export function sortByLength(array) {
    return array.slice(0).sort(function(a, b) {
        // ASC  -> a.length - b.length
        // DESC -> b.length - a.length
        if (!defined(a)) {
            a = '';
        }
        if (!defined(b)) {
            b = '';
        }
        return b.length - a.length;
    });
}
utils.sortByLength = sortByLength;

/*
type - integer | string
 */
export function sortByIndices(array, type) {
    var len = array.length;
    var indices = new Array(len);
    for (var i = 0; i < len; ++i) {
        indices[i] = i;
    }
    indices.sort(function(i, j) {
        // ASC  -> a.length - b.length
        // DESC -> b.length - a.length
        var a = array[i],
            b = array[j];
        if (!defined(a)) {
            a = '';
        }
        if (!defined(b)) {
            b = '';
        }
        if (type == 'integer') {
            return b > a ? 1 : b < a ? -1 : 0;
        }
        return b.length > a.length ? 1 : b.length < a.length ? -1 : 0;
    });
    return indices;
}
utils.sortByIndices = sortByIndices;

export function isHidden(el) {
    return el.offsetParent === null;
}
utils.isHidden = isHidden;

export function each(obj, fn) {
    if (typeof obj !== 'object') {
        return false;
    }

    if (isArray(obj) || typeof obj.length === 'number') {
        Array.prototype.forEach.call(obj, fn);
    } else {
        for (var attr in obj) {
            if (obj.hasOwnProperty(attr)) {
                fn(obj[attr], attr, obj);
            }
        }
    }
}

utils.each = each;

/**
 * Extend one or more objects with the members of another
 * @param target
 * @returns {Object}
 */

export function extend(target) {
    if (!arguments[1]) {
        return;
    }
    for (var ii = 1, ll = arguments.length; ii < ll; ii++) {
        var source = arguments[ii];
        for (var prop in source) {
            if (!target[prop] && source.hasOwnProperty(prop)) {
                target[prop] = source[prop];
            }
        }
    }
    return target;
}
utils.extend = extend;

/**
 * Inherits an Object
 * @param {Object|*} subClass
 * @param {Object|*} superClass
 */

export function inherit(subClass, superClass) {
    if (typeof superClass !== 'function' && superClass !== null) {
        throw new TypeError(`Super expression must either be null or a function, not ${typeof superClass}`);
    }
    var subClassProto = subClass.prototype;
    subClass.prototype = Object.create(superClass && superClass.prototype, {
        constructor: {
            value: subClass,
            enumerable: false,
            writable: true,
            configurable: true
        }
    });
    if (superClass) {
        Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : (subClass.__proto__ = superClass);
    }
    // restore previous methods
    extend(subClass.prototype, subClassProto);
}
utils.inherit = inherit;

/**
 *
 * @param {Array} array
 * @param {*} element
 * @returns {Array}
 */
export function removeElementFromArray(array, element) {
    var index = array.indexOf(element);
    if (index > -1) {
        array.splice(index, 1);
    }
    return array;
}
utils.removeElementFromArray = removeElementFromArray;

/**
 *
 * Get the computed CSS value for given element and property, only for numerical
 * properties. For width and height, the dimension of the inner box (excluding
 * padding) is returned. Used for fitting the chart within the container.
 *
 * @param {HTMLElement} el - A HTML element.
 * @param {String} prop - The property name.
 * @param {Boolean} [convertToInt=true] - Parse to integer.
 * @returns {Number|*}
 */

export function getStyle(el, prop, convertToInt) {
    var style;

    if (el.isD3Selection) {
        el = el.node();
    }

    // Get the computed style
    style = window.getComputedStyle(el);

    // For width and height, return the actual inner pixel size
    if (prop === 'width') {
        return consts.mathMin(el.offsetWidth, el.scrollWidth) - parseComputedStyle('padding-left', style, convertToInt) - parseComputedStyle('padding-right', style, convertToInt);
    } else if (prop === 'height') {
        return consts.mathMin(el.offsetHeight, el.scrollHeight) - parseComputedStyle('padding-top', style, convertToInt) - parseComputedStyle('padding-bottom', style, convertToInt);
    }

    

    if (style) {
        style = style.getPropertyValue(prop);
        if (pick(convertToInt, prop !== 'opacity')) {
            style = pInt(style);
        }
    }
    return style;
}

utils.getStyle = getStyle;

function parseComputedStyle(prop, style, convertToInt){
    let value;
    if (style) {
        value = style.getPropertyValue(prop);
        if (pick(convertToInt, prop !== 'opacity')) {
            value = pInt(value);
        }
    }
    return value;
}

/**
 *
 * @param element
 * @returns {boolean}
 */
export function isHTMLElement(element) {
    return element instanceof HTMLElement;
}
utils.isHTMLElement = isHTMLElement;

/**
 * Check if an element is an array, and if not, make it into an array.
 */

export function splat(obj) {
    return isArray(obj) ? obj : [obj];
}
utils.splat = splat;

export function parseShortHandValue(value) {
    var t, r, b, l;
    if (defined(value)) {
        var parsedValue = value.toString().split(' ');

        t = pInt(priorityBasedValue([parsedValue[0], 0]));
        r = pInt(priorityBasedValue([parsedValue[1], parsedValue[0], 0]));
        b = pInt(priorityBasedValue([parsedValue[2], parsedValue[0], 0]));
        l = pInt(priorityBasedValue([parsedValue[3], parsedValue[1], parsedValue[0], 0]));
    } else {
        t = 0;
        r = 0;
        b = 0;
        l = 0;
    }

    return {
        top: t,
        right: r,
        bottom: b,
        left: l
    };
}
utils.parseShortHandValue = parseShortHandValue;

export function getBorderRadius(radius) {
    if (isObject(radius)) {
        return radius;
    }
    var parsedVal = {};
    var splittedStr = radius != null ? radius.toString().split('/') : null;
    var rx = splittedStr != null && splittedStr[0] != null ? splittedStr[0].split(' ') : [],
        ry = splittedStr != null && splittedStr[1] != null ? splittedStr[1].split(' ') : rx;

    /*
     * index 0-->topleft,  1-->topright,  2-->bottomright,  3-->bottomleft
     */
    parsedVal.tlx = pInt(priorityBasedValue([rx[0]]));
    parsedVal.tly = pInt(priorityBasedValue([ry[0], rx[0]]));

    parsedVal.trx = pInt(priorityBasedValue([rx[1], rx[0]]));
    parsedVal.trY = pInt(priorityBasedValue([ry[1], ry[0], rx[1], rx[0]]));

    parsedVal.brx = pInt(priorityBasedValue([rx[2], rx[0]]));
    parsedVal.bry = pInt(priorityBasedValue([ry[2], ry[0], rx[2], rx[0]]));

    parsedVal.blx = pInt(priorityBasedValue([rx[3], rx[1], rx[0]]));
    parsedVal.bly = pInt(priorityBasedValue([ry[3], ry[1], rx[3], rx[0]]));

    return parsedVal;
}
utils.getBorderRadius = getBorderRadius;

/**
 *
 * @param priorities
 * @param defaultValue
 * @return {number|*}
 */

export function priorityBasedValue(priorities, defaultValue) {
    var value = null,
        size = priorities.length,
        i = 0;

    defaultValue = pick(defaultValue, 0);
    while (i < size && !defined(value)) {
        value = priorities[i] != null ? priorities[i] : null;
        i++;
    }
    return pick(value, defaultValue);
}
utils.priorityBasedValue = priorityBasedValue;

/**
 *
 * @param {String} string
 * @return {String}
 */
export function toCapitalize(string) {
    return string.charAt(0).toUpperCase() + string.slice(1);
}

utils.toCapitalize = toCapitalize;

/**
 *
 * @param {CanvasRenderingContext2D} context
 */
export function fixHDPI(context) {
    var canvas = context.canvas,
        pixelRatio = DEVICE_PIXEL_RATIO;
    canvas.style.width = `${canvas.width}px`;
    canvas.style.height = `${canvas.height}px`;

    canvas.setAttribute('width', canvas.width * pixelRatio);
    canvas.setAttribute('height', canvas.height * pixelRatio);
    context.scale(pixelRatio, pixelRatio);
    return context;
}

utils.fixHDPI = fixHDPI;

/**
 *
 * @param {Array} array
 * @param {function} callback
 * @return {Array}
 */
export function arrayFind(array, callback) {
    if (Array.prototype.find) {
        return array.find(callback);
    } else {
        // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find
        var len = array.length >>> 0,
            i = 0;

        while (i < len) {
            var kValue = array[i];
            if (callback.call(callback, kValue, i, array)) {
                return kValue;
            }
            i++;
        }
        return undefined;
    }
}

utils.arrayFind = arrayFind;

function moveArrayElement(array, from, to) {
    array.splice(to, 0, array.splice(from, 1)[0]);
}

utils.moveArrayElement = moveArrayElement;

/**
 *
 * @param node
 * @return {*}
 */
export function getRootNode(node) {
    if (node?.getRootNode) {
        return node.getRootNode();
    } else if (node?.parentNode != null) {
        return getRootNode(node.parentNode);
    }
    return node;
}
utils.getRootNode = getRootNode;

/**
 *
 * @param element
 * @return {boolean}
 */
function isShadowElement(element) {
    return isShadowRoot(getRootNode(element));
}

utils.isShadowElement = isShadowElement;

/**
 *
 * @param node
 * @return {boolean}
 */
function isShadowRoot(node) {
    return node.nodeName === '#document-fragment' && node.constructor.name === 'ShadowRoot';
}

utils.isShadowRoot = isShadowRoot;

/**
 *
 * @param element
 * @return {boolean}
 */
export function isElementOnDocument(element) {
    return isShadowElement(element) || document.body.contains(element);
    // if (isShadowElement(element)) {
    //     return d3.select(element).style("display") !== NONE;
    // } else {
    //     return document.body.contains(element)
    // }
}
utils.isElementOnDocument = isElementOnDocument;

/**
 *
 * @param element
 * @return {boolean}
 */
export function isElementVisible(element) {
    return isElementOnDocument(element) && d3_select(element).style('display') !== consts.NONE;
}

utils.isElementVisible = isElementVisible;

/**
 *
 * @param node
 * @return {*}
 */
export function getRootBody(node) {
    var rootNode = getRootNode(node);
    if (isShadowRoot(rootNode)) {
        var fakeBodySelector = '_zc_fakeBody',
            fakeBody = rootNode.querySelector(`.${fakeBodySelector}`);

        if (!fakeBody) {
            fakeBody = document.createElement('div');
            fakeBody.className = fakeBodySelector;
            rootNode.appendChild(fakeBody);
        }
        return fakeBody;
    } else {
        return document.body;
    }
}

utils.getRootBody = getRootNode;

/**
 *
 * @param element
 * @return {number}
 */
export function getDocumentIndex(element) {
    return getRootNode(element)?.ZC_DOCUMENT_INDEX;
}
utils.getDocumentIndex = getDocumentIndex;

/**
 *
 * @param {d3.selection|HTMLElement}element
 * @return {*|ClientRect|DOMRect}
 */
export function getBoundingRect(element) {
    if (element.isD3Selection) {
        element = element.node();
    }

    if (defined(element) && isElementVisible(element)) {
        return element.getBoundingClientRect();
    }

    var t = {};
    t.x = t.y = t.width = t.height = t.top = t.left = t.right = t.bottom = 0;
    return t;
}
utils.getBoundingRect = getBoundingRect;

export function perToPx(val, dim, defaultVal) {
    return typeof val === 'string' && val.indexOf('%') > -1 ? dim * +val.replace(/%/g, '') * 0.01 : validateNullVal(val, defaultVal);
}
utils.perToPx = perToPx;

/**
 *
 * @param obj
 * @return {any}
 */
export function _super(obj) {
    return Object.getPrototypeOf(obj.constructor.prototype);
}
utils._super = _super;

/**
 *
 * @param {Array} array
 * @param {any} item
 * @return {boolean}
 */
export function arrayIncludes(array, item) {
    if (!defined(array)) {
        return false;
    }
    if (array.includes) {
        return array.includes(item);
    } else {
        return array.indexOf(item) > -1;
    }
}
utils.arrayIncludes = arrayIncludes;

/**
 *
 * @type {arrayIncludes}
 */
export const stringIncludes = arrayIncludes;

/**
 *
 * @param number
 * @param minMax
 * @return {boolean}
 */
export function numberIsBetween(number, minMax) {
    var min = Math.min.apply(Math, minMax),
        max = Math.max.apply(Math, minMax);
    return number >= min && number <= max;
}
utils.numberIsBetween = numberIsBetween;

/**
 *
 * @param property
 * @return {boolean}
 */
export function isStylePropertySupported(property) {
    return property in document.documentElement.style;
}
utils.isStylePropertySupported = isStylePropertySupported;

/**
 *
 * @param obj
 * @return {boolean}
 */
export function isEmptyObject(obj) {
    return !defined(obj) || (Object.keys(obj).length === 0 && obj.constructor === Object);
}
utils.isEmptyObject = isEmptyObject;

/**
 *
 * @param {[[number, number], [number, number]]} bound
 * @return {{x: *, width: number, y: *, height: number}}
 */
export function boundToRect(bound) {
    return {
        x: bound[0][0],
        y: bound[0][1],
        width: mathAbs(bound[1][0] - bound[0][0]),
        height: mathAbs(bound[1][1] - bound[0][1])
    };
}
utils.boundToRect = boundToRect;

/**
 * returns distinct values in an array
 * @param {Array} array
 * @return {Array}
 */
export function getDistinct(array, accessor) {
    var values = [...new Set(array).values()];

    if (accessor) {
        values = values.map(accessor);
    }

    return values;
}
utils.getDistinct = getDistinct;

/**
 * @param  {Object} obj [description]
 * @return {Array}     [description]
 */
export function objectValues(obj) {
    if (Object.values) {
        return Object.values(obj);
    }

    return Object.keys(obj).map(key => obj[key]);
}
utils.objectValues = objectValues;

/**
 *
 * @return {number}
 */
export const getScrollbarWidth = (() => {
    // Creating invisible container
    let scrollbarWidth = null;
    return () => {
        if (defined(scrollbarWidth)) {
            return scrollbarWidth;
        }

        let outer = d3_select(document.body || document.documentElement)
            .append('div')
            .styles({
                width: '100%',
                visibility: 'hidden',
                overflow: 'scroll',
                '-ms-overflow-style': 'scrollbar'
            });

        let inner = outer.append('div').styles({
            width: '100%'
        });

        // Calculating difference between container's full width and the child width
        scrollbarWidth = outer.node().offsetWidth - inner.node().offsetWidth;

        // Removing temporary elements from the DOM
        outer.remove();

        return scrollbarWidth;
    };
})();

utils.getScrollbarWidth = getScrollbarWidth;

/**
 *
 * @param element
 * @return {boolean}
 */
export const elementIsScrollable = (element, measureName = 'height') => {
    if (element.isD3Selection) {
        element = element.node();
    }
    return measureName === 'width' ? element.scrollWidth > element.offsetWidth : element.scrollHeight > element.offsetHeight;
};

utils.elementIsScrollable = elementIsScrollable;

/*
accessor - callback function to acccess the key
a - array of data
x - search value
returns array of indices where the search value present in the given array
 */
export function searchIndex(accessor) {
    return function(a, x) {
        var _i = [];
        a.map(function(d, i) {
            if (accessor(d) == x) {
                _i.push(i);
            }
        });
        return _i;
    };
}
utils.searchIndex = searchIndex;

export function objectEntries(obj) {
    if (Object.entries) {
        return Object.entries(obj);
    } else {
        // Polyfill
        var ownProps = Object.keys(obj),
            i = ownProps.length,
            resArray = new Array(i); // preallocate the Array

        while (i--) {
            resArray[i] = [ownProps[i], obj[ownProps[i]]];
        }

        return resArray;
    }
}
utils.objectEntries = objectEntries;

export function firstAndLastItem(array) {
    if (isArray(array)) {
        return [array[0], array[array.length - 1]];
    }
    return array;
}
utils.firstAndLastItem = firstAndLastItem;

/**
 *
 * @param xs
 * @param key
 * @return {*}
 */
export function groupBy(xs, key) {
    return xs.reduce(function(rv, x) {
        (rv[x[key]] = rv[x[key]] || []).push(x);
        return rv;
    }, {});
}

utils.groupBy = groupBy;

/**
 *
 * @param obj
 * @param except
 */
export function destroyObjectProperties(obj, except) {
    /**
     *
     * @param val
     * @param n
     */
    function fn(val, n) {
        // If the object is non-null and destroy is defined

        if (val && val !== except) {
            // Invoke the remove
            val.remove && val.remove();
            // Invoke the destroy
            val.destroy && val.destroy();
        }

        // remove the timeouts
        // if (isNumber(val)) {
        // clearTimeout(val);
        // clearInterval(val);
        // }

        // Delete the property from the object.
        obj[n] = null;
        delete obj[n];
    }

    for (var key in obj) {
        if (Object.hasOwnProperty.call(obj, key)) {
            fn.call(obj[key], obj[key], key, obj);
        }
    }
}

utils.destroyObjectProperties = destroyObjectProperties;

/**
 * 
 * @param {*} obj 
 * @param {Array} list - list of items to include/exclude
 * @param {boolean} exclude - toggle the list of items from remove to exclude
 */
export function deleteProperties(obj = {}, list = [], exclude = false) {
    exclude ? (list = Object.keys(obj).filter(key => list.includes(key))) : 0;
    list.forEach(item => delete obj[item]);
}

/**
 *
 * @param partialValue
 * @param totalValue
 * @return {number}
 */
export function calculatePercentage(partialValue, totalValue) {
    return (100 * partialValue) / totalValue;
}

export function getDepth(level) {
    return level && level.split('_').length;
}
utils.getDepth = getDepth;

export function getLevel(d) {
    return d.level || (d.data && d.data.level);
}
utils.getLevel = getLevel;

export function hasSameParent(level, parent, layer = 0, increment = 1) {
    parent = parent && parent.split('_') || [];
    level = level && level.split('_') || [];
    let flag = false;

    if (parent.length && level.length) {
        while(layer < parent.length && layer > -1) {
            if (!(flag = parent[layer] === level[layer])) {return flag;}
            layer = layer + increment;
        }
    }
    return flag;
}
utils.hasSameParent = hasSameParent;

export function isAncestorDisabled(data, d) {
    const seriesIndices = d.level ? d.level.split('_') : [];
    for (let i = 0; i < seriesIndices.length; i++) {
        data = data[seriesIndices[i]];
        if (data.disabled) {
            return true;
        }
    }
    return false;
}
utils.isAncestorDisabled = isAncestorDisabled
