// @zohocharts/zc/src/polaraxis/polarutils.js
//$Id$
//polarutils.js

import { Registry } from '../core/zc-registry';
import { mathPI, mathCos, mathSin, mathMax, mathMin } from '../core/zc-constants';
import { validateNullVal, defined, perToPx } from '../utils/zcutils';
import { d3_scaleLinear } from '../core/d3-constants';

export function polarToXY(r, a) {
    a = a - mathPI / 2;
    return [mathCos(a) * r, mathSin(a) * r];
}
Registry.setComponent('geometryUtils.polarToCartesian', polarToXY);

export function d0_360(a) {
    var a = a % (2 * mathPI);
    return a < 0 ? a + 2 * mathPI : a;
}

export function polarRefPoint(bb, w, h, r, SW) {
    var leftWidth = 0 - (bb.x - 0.5 * SW),
        rightWidth = bb.width + SW - leftWidth,
        topHeight = 0 - (bb.y - 0.5 * SW),
        bottomHeight = bb.height + SW - topHeight;
    var leftWidthRemaining = r - leftWidth,
        rightWidthRemaining = r - rightWidth,
        topHeightRemaining = r - topHeight,
        bottomHeightRemaining = r - bottomHeight;
    return [w / 2 - leftWidthRemaining / 2 + rightWidthRemaining / 2, h / 2 - topHeightRemaining / 2 + bottomHeightRemaining / 2];
}

export function getDY(angle) {
    return d3_scaleLinear()
        .range([0, 1])
        .domain([angle < mathPI ? mathPI / 2 : (3 * mathPI) / 2, mathPI])(angle);
}
Registry.setComponent('geometryUtils.dyFromAngle', getDY);

export function isAngleInside(theta1, theta2, angle) {
    if (theta2 < theta1) {
        //Swap the start and end angle if endangle is less than start angle
        theta2 = theta1 + ((theta1 = theta2) - theta2);
    }
    if (theta1 < 0 && theta2 < 0) {
        (theta1 = theta1 + 2 * mathPI), (theta2 = theta2 + 2 * mathPI);
    }
    if (theta1 < 0 && theta2 >= 0) {
        if (angle < 0) {
            theta2 = 2 * mathPI;
        } else {
            theta1 = 0;
        }
    }
    angle = angle < 0 ? angle + mathPI * 2 : angle; //Mapping [-Math.PI,Math.PI] to [0,2*Math.PI]
    return theta1 <= angle && angle <= theta2;
}
Registry.setComponent('geometryUtils.isAngleInside', isAngleInside);

export function radianToDegree(rad) {
    return (rad * 180) / mathPI;
}
export function degreeToRadian(degree) {
    return degree / 180 * mathPI;
}

/*
 * Returns the inner radius for pie | dial chart. If the inner radius is in percentage. It converts to pixel and considers bandWidth & maxBandWidth
 * @param {Object} commonPlotOptions
 * @param {String} | {Number} rad
 * @return {Number} inner radius
 */
export function getInnerRadius(commonPlotOptions, rad) {
    var hasCustomBandWidth =
            defined(commonPlotOptions.bandWidth) || (commonPlotOptions.levelMarker && defined(commonPlotOptions.levelMarker.bandWidth)),
        hasCustomMaxBandWidth =
            defined(commonPlotOptions.maxBandWidth) ||
            (commonPlotOptions.levelMarker && defined(commonPlotOptions.levelMarker.maxBandWidth));
    var innerRadius = mathMax(0, perToPx(commonPlotOptions.innerRadius, rad, 0));
    if (hasCustomBandWidth || hasCustomMaxBandWidth) {
        var bandWidth = defined(commonPlotOptions.bandWidth) ? perToPx(commonPlotOptions.bandWidth, rad, 0) : rad - innerRadius,
            levelMarkerBandWidth = defined(commonPlotOptions.levelMarker.bandWidth)
                ? perToPx(commonPlotOptions.levelMarker.bandWidth, rad, 0)
                : rad - innerRadius,
            maxBandWidth = validateNullVal(commonPlotOptions.maxBandWidth, rad),
            maxLevelMarkerBandWidth = validateNullVal(commonPlotOptions.levelMarker.maxBandWidth, rad);

        bandWidth = mathMin(bandWidth, maxBandWidth);
        levelMarkerBandWidth = mathMin(levelMarkerBandWidth, maxLevelMarkerBandWidth);

        return rad - mathMax(bandWidth, levelMarkerBandWidth);
    } else {
        // #ZC1724
        return mathMin(innerRadius, rad);
    }
}
