// @zohocharts/zc/src/polaraxis/polaraxis-helpers.js
//$Id$
//polar-axis-helpers.js
import { mathPI } from '../core/zc-constants';
import { validateNullVal } from '../utils/zcutils';
import { getFormattedValue } from '../formatter/formatter';
import { isFacet } from '../axis/hierarchy-axis-utils';
import { Registry } from '../core/zc-registry';

export function getAxisObject(chartObj, type, index = 0) {
    return getAxisObjects(chartObj, type)[index];
}

export function getAxisObjects(chartObj, type) {
    let axes = chartObj.axes[type];

    if (isFacet(chartObj)) {
        if (type === 'y') {
            let hierarchical = chartObj.dataObject.hierarchical,
                nodeList = hierarchical.pivotLayout.getRootNodeList('row'),
                nodeListCount = nodeList.getCount(),
                hasMultiCluster = nodeList.hasMultiCluster();

            if (hasMultiCluster || nodeListCount) {
                return axes.leaves;
            } else {
                return axes.groups;
            }
        } else {
            let hasSplitLeavesColumns = chartObj.systemConf.chart.axes.xaxis.splitLeaves;
            if (hasSplitLeavesColumns) {
                return axes.leaves;
            } else {
                return [{ scale: axes.leafScaleGrid }];
            }
        }
    } else {
        return type === 'y' ? axes.groups : [{ scale: axes.scale }];
    }
}

export function getAxisData(chartObj, type, index = 0) {
    let axesConfig = chartObj.systemConf.chart.axes;
    return type == 'x' ? axesConfig.xaxis : axesConfig.yaxis[index];
}

export function getPlotConfig(chartObj) {
    let chartName = chartObj.chartTypes[0].name;
    return chartObj.systemConf.chart.plot.plotoptions[chartName];
}

export function getFormat(chartObj, type, index, _class) {
    let column = chartObj.dataset.getColumn(type, index),
        { datatype } = column,
        formatConf = column[datatype];

    return d => getFormattedValue(formatConf, d, datatype, chartObj, _class);
}

export function getPolarRange(chartObj, type, config) {
    let polarConfig = config || getPolarRangeConfig(chartObj, type),
        { reversed, startAngle, endAngle } = polarConfig;

    return reversed
        ? [validateNullVal(endAngle, 2 * mathPI), validateNullVal(startAngle, 0)]
        : [validateNullVal(startAngle, 0), validateNullVal(endAngle, 2 * mathPI)];
}
Registry.setComponent('axis.polar.helpers.getPolarRange', getPolarRange);

export function getPolarRangeConfig(chartObj, type) {
    let commonPlotOptions = getPlotConfig(chartObj),
        axisData = getAxisData(chartObj, type);
    return {
        reversed: axisData.reversed,
        startAngle: commonPlotOptions.startAngle,
        endAngle: commonPlotOptions.endAngle
    };
}
Registry.setComponent('axis.polar.helpers.getPolarRangeConfig', getPolarRangeConfig);
