// @zohocharts/zc/src/polaraxis/polar-ticklabel-overlap.js
//
import { TEXT_ELEMENT } from '../core/element-constants';
import { mathPI, mathMax } from '../core/zc-constants';
import { d3_select } from '../core/d3-constants';
import { bbox } from '../utils/zcutils';
import { doOverlap } from '../utils/geometry-utils';
import { getSvgTransform } from '../utils/svg-transforms';

function PolarTickOverlap(wrap, scale) {
    let tickLabels = wrap.selectAll(`${TEXT_ELEMENT}.ticklabel`),
        { right, left } = getPartition(tickLabels, scale),
        maxJumps = mathMax(findJumps(right), findJumps(left)),
        jump = 0;

    if (!maxJumps) {
        return;
    }

    tickLabels.each(function() {
        if (jump && jump <= maxJumps) {
            d3_select(this).style('display', 'none');
        }
        jump = jump === maxJumps ? 0 : jump + 1;
    });

    let visibleTickLabels = tickLabels.filter(function() {
        return d3_select(this).style('display') !== 'none';
    });

    ({ right, left } = getPartition(visibleTickLabels, scale));
    maxJumps = mathMax(findJumps(right), findJumps(left));

    if (!maxJumps) {
        return;
    }

    handleOverlap(right);
    handleOverlap(left);
}

function getPartition(tickLabels, scale) {
    let right = [],
        left = [];

    tickLabels.each(function(d) {
        let angle = scale(d),
            ele = d3_select(this),
            info = { bb: bbox(this), xy: getSvgTransform(this.parentNode).translate, ele };

        if (angle === 0 || angle === mathPI || angle === mathPI * 2) {
            right.push(info);
            left.push(info);
        } else if (angle > 0 && angle < mathPI) {
            right.push(info);
        } else {
            left.push(info);
        }
    });

    return { left, right };
}

function findJumps(arr) {
    let maxJumps = 0;
    arr.sort((a, b) => a.xy[1] - b.xy[1]);

    for (var i = 0; i < arr.length - 1; i++) {
        let reached = false,
            jumps = 0,
            refRect = createRect(arr[i]);
        for (var j = i + 1; j < arr.length && !reached; j++) {
            let curRect = createRect(arr[j]);

            if (!doOverlap(refRect, curRect)) {
                reached = true;
            } else {
                jumps++;
            }
        }

        maxJumps = mathMax(maxJumps, jumps);
    }
    return maxJumps;
}

function handleOverlap(arr) {
    for (var i = 0; i < arr.length - 1; i++) {
        if (arr[i].hidden) {
            return;
        }

        let reached = false,
            refRect = createRect(arr[i]);

        for (var j = i + 1; j < arr.length && !reached; j++) {
            let { ele } = arr[j],
                curRect = createRect(arr[j]);

            if (doOverlap(refRect, curRect)) {
                arr[j].hidden = true;
                ele.style('display', 'none');
            } else {
                reached = true;
            }
        }
    }
}

function createRect(obj) {
    let { bb, xy } = obj;
    return {
        x1: bb.x + xy[0],
        y1: bb.y + xy[1],
        x2: bb.x + bb.width + xy[0],
        y2: bb.y + bb.height + xy[1]
    };
}

export { PolarTickOverlap };
