// @zohocharts/zc/src/polaraxis/polaraxis-init.js
//$Id$
//polaraxis-init.js

import { LabelRegistry } from '../label/label-registry';
import { PolarTickHandler } from './polar-ticklabel-overflow';
import { PolarTickOverlap } from './polar-ticklabel-overlap';
import { axis, scale, charttype, renderer_helpers } from '../core/main-zc-init';
import { dimension_based_intelligence } from '../utils/intelligence-utils';
import { validateNullVal, utils, bbox, perToPx, pick, isArray, invokeFunction } from '../utils/zcutils';
import { getInnerRadius, polarToXY, polarRefPoint } from './polarutils';
import { appendEle, createElement, addSVGtitleEle } from '../utils/zcsvgutils';
import { mathMin, mathMax, mathPow, mathSQRT, NONE } from '../core/zc-constants';
import { d3_arc, d3_min, d3_max, d3_select } from '../core/d3-constants';
import { TEXT_ELEMENT, PATH_ELEMENT } from '../core/element-constants';
import { LEFT_ALIGN, OUTER_ALIGN, BOTTOM_ALIGN, INNER_ALIGN, RIGHT_ALIGN } from '../core/align-constants';
import { merged_axis_column_intersect, getAxisData } from '../axis/axisutils';
import { isDatalabelEnabled } from '../label/datalabelsUtils';
import { dialUtils } from '../dial/dial-utils';
import { $ZCG } from '../core/zcglobal';
// import { createSVGRenderer } from '../plot/svgRenderer';
import { plot } from '../core/main-zc-init';
import { translateString, getSvgTransform } from '../utils/svg-transforms';
import { getAxisObject, getAxisObjects, getFormat, getPolarRange } from './polaraxis-helpers';
import { minMaxLabel } from './components/minmaxlabel';
import { MinmaxLabelSpec } from './components/minmaxlabel-spec';
import { getGridList } from '../core/data-processor/pivot-layout-helpers';
import { isFacet, getAxisDataIndex } from '../axis/hierarchy-axis-utils';
import { getPolarTickValues } from './polar-ticks';

export * from './polaraxisComponent';

axis.polar.render = function(chartObj) {
    let types = chartObj.dataObject.isPolarAxisCategory ? ['x', 'y'] : ['y'],
        { plotarea } = chartObj,
        facetRendering = isFacet(chartObj);

    chartObj.plotarea_approx = null;
    types.forEach(type => {
        let wrap = createElement('g', `.${type}-axis`, [1], chartObj.wrap.selectAll('g.chartgroup'));
        setRange4Scale(chartObj, type);

        if (facetRendering) {
            var gridList = getGridList(chartObj.dataObject);
            chartObj.cache.polar = chartObj.cache.polar || {};

            gridList.forEach(grid => {
                let { row, column } = grid,
                    axisOrder = type === 'x' ? 0 : row,
                    axisObj = getAxisObject(chartObj, type, axisOrder),
                    bound = renderer_helpers.getBound(chartObj, grid.data[0]),
                    axisDataOrder = type === 'x' ? 0 : getAxisDataIndex(type, axisOrder, chartObj);

                let axisWrap = createElement('g', `.${type}-axis-${row}-${column}`, [1], wrap);
                transformWrap(axisWrap, [plotarea.left + bound.x, plotarea.top + bound.y]);

                drawAxisComp(chartObj, axisObj, type, axisDataOrder, axisWrap, bound, chartObj.plotarea_approx);

                chartObj.cache.polar[`${row}-${column}`] = bound.polar;
            });
        } else {
            let axisObj = getAxisObject(chartObj, type),
                bound = renderer_helpers.getBound(chartObj);

            let axisWrap = createElement('g', `.${type}-axis-0`, [1], wrap);
            transformWrap(axisWrap, [plotarea.left, plotarea.top]);

            drawAxisComp(chartObj, axisObj, type, 0, axisWrap, bound);
        }
    });
};

function drawAxisComp(chartObj, axisObj, type = 'y', index = 0, container, bound, polarMeta) {
    let mode = getMode(chartObj, type),
        comp = mode === 'polar' ? drawPolarAxisComp : drawAngularAxisComp;

    comp(chartObj, axisObj, type, index, container, bound, polarMeta);
}

function drawPolarAxisComp(chartObj, axisObj, type = 'y', index = 0, container, bound, polarMeta) {
    var systemconf = chartObj.systemConf,
        renderer = chartObj.renderer[0],
        plotarea = bound,
        chart = chartObj.chartTypes[0],
        cType = chart.value,
        chartName = chart.name;

    var commonPlotOptions = systemconf.chart.plot.plotoptions[chartName];
    var axis_data = getAxisData(chartObj, type, index),
        outerPadding = +commonPlotOptions.outerPadding || 0,
        maxRadius = +commonPlotOptions.maxRadius || 0;
    var isDial = cType === charttype.dial;

    var show = isAxis(chartObj, type, axis_data);
    var axislineSize = show && axis_data.axisline.show ? validateNullVal(axis_data.axisline.strokeWidth, 1) : 0,
        ticklabel = axis_data.ticklabel,
        ticklabelShowAs = ticklabel.showAs,
        ticklabelPad = validateNullVal(axis_data.ticklabel.marginLeft, 5),
        tickmarkSize = validateNullVal(axis_data.tickmark.size, 3),
        subtickmarkSize = axis_data.subtickmark ? validateNullVal(axis_data.subtickmark.size, 2) : 0,
        ticklabelFontsize = utils.getVal(
            axis_data.ticklabel.fontSize || systemconf.chart.fontSize || systemconf.canvas.fontSize || $ZCG.fontSize
        ),
        orient = axis_data.orient || OUTER_ALIGN,
        tickOrient = axis_data.tickOrient,
        textOverlap = axis_data.ticklabel.textOverlap || 'hidden',
        _childNodes = ele =>
            ele.selectAll(function() {
                return this.childNodes;
            });

    var w = plotarea.width - 2 * outerPadding,
        h = plotarea.height - 2 * outerPadding,
        radius = pick(polarMeta && polarMeta.radius, mathMin(w, h) / 2),
        innerRadius = pick(polarMeta && polarMeta.innerRadius, getInnerRadius(commonPlotOptions, radius)),
        padding = show ? +commonPlotOptions.padding || 0 : 0;
    chartObj.plotarea_approx = { x: 0, y: 0, innerRadius: innerRadius };

    var scaleObj = axisObj.scale;
    var range = scaleObj.range();

    var columnIdx = chartObj.dataset.getColumnIdx(type, index),
        columnInfo = chartObj.dataset.getColumn(type, index),
        dataObject = chartObj.dataObject,
        categories =
            dataObject.isPolarAxisCategory &&
            dataObject.categoryHierarchy &&
            dataObject.hierarchical.pivotLayout.columns.count &&
            !dataObject.hierarchical.pivotLayout.columns.splitColumnLeaves &&
            dataObject.getCategories(type, index),
        tickValues = getPolarTickValues({ chartObj, type, index, config: axis_data, scale: scaleObj, radius, plotarea, container });
    scaleObj.range(range); // Resetting range that is changed in getPolarTickValues

    var axisRadius = pick(
            polarMeta && polarMeta.axisRadius,
            orient == INNER_ALIGN
                ? tickOrient === OUTER_ALIGN
                    ? innerRadius - padding - axislineSize / 2 - mathMax(subtickmarkSize, tickmarkSize + ticklabelPad + ticklabelFontsize)
                    : innerRadius
                : radius
        ),
        tickFormat = (d, i) => {
            let val = categories ? categories[d] : d; // ZC2682
            let formattedVal = ticklabelShowAs ? invokeFunction(ticklabelShowAs, val, ticklabel, { data: [columnInfo], index: i }, chartObj) 
                : getFormat(chartObj, type, index, 'axis')(val, i);
            return formattedVal;
        };

    var axisComponent = axis
        .polar()
        .scale(scaleObj)
        .chartObj(chartObj)
        .radius(axisRadius)
        .columnIndices([columnIdx])
        .tickFormat(tickFormat)
        .tickValues(tickValues)
        .type(type);

    _childNodes(container).attr('transform', null);
    var wrap = createElement('g', '.axisgroup', show ? [axis_data] : [], container),
        getTickHandler = () => {
            let args = [wrap, w / 2, axis_data.ticklabel, chartObj, scaleObj, axisComponent];
            return PolarTickHandler(...args);
        },
        addTooltip = function() {
            if (axis_data.ticklabel.tooltip === true) {
                let ticklabelEle = wrap.selectAll(TEXT_ELEMENT);
                ticklabelEle.each(function(d) {
                    addSVGtitleEle(d3_select(this), tickFormat(d));
                });
            }
        },
        renderAxis = function() {
            wrap.selectAll(`${TEXT_ELEMENT}.ticklabel`).style('display', null);
            wrap.call(axisComponent)
                .call(getTickHandler())
                .call(() => (textOverlap === 'hidden' ? PolarTickOverlap(wrap, scaleObj) : null))
                .call(addTooltip);
        };

    renderAxis();
    container.style('visibility', 'hidden');

    if (isDial) {
        drawMinMaxLabel(radius);
    }

    if (!polarMeta) {
        //dummy arc
        var arc = d3_arc()
            .innerRadius(
                !isDial ||
                    commonPlotOptions.needle.show ||
                    (chartObj.datalabels.moduleExists && isDatalabelEnabled(commonPlotOptions.datalabels, chartObj))
                    ? 0
                    : innerRadius
            )
            .outerRadius(radius)
            .startAngle(range[0])
            .endAngle(range[1]);
        var dummyArc = appendEle(container, PATH_ELEMENT, [1], '.', 'class', 'dummyArc').attr('d', arc(1));

        //Minmax label and datalabel - dial chart
        var datalabel = createElement('g', '.datalabel', [1], container);
        var targetMarkerSeries = createElement('g', '.targetSeries', [1], container);
        let gridList;
        if (isDial) {
            let hierarchical = chartObj.dataObject.hierarchical,
                facetRendering = hierarchical && hierarchical.facetRendering;

            if (facetRendering) {
                gridList = getGridList(chartObj.dataObject);
            } else {
                gridList = [{ data: chartObj.seriesdata }];
            }

            if (!renderer) {
                var curSeriesidx = 0;
                renderer = chartObj.renderer[curSeriesidx] = plot.svgRenderer.createSVGRenderer(chartName, chartObj);
                chartObj.rendererConf = [];
                chartObj.rendererConf[curSeriesidx] = new plot.rendererConf(); // TODO: PIVOT
                renderer_helpers.setRendererProp(
                    chartObj,
                    renderer,
                    gridList[0].data,
                    curSeriesidx,
                    curSeriesidx,
                    [],
                    curSeriesidx,
                    null,
                    curSeriesidx
                );
            }

            renderer.firedList = {};
            let dialLabelRenderer = LabelRegistry.get('dial');
            if (!dialLabelRenderer._empty) {
                datalabel.call(
                    dialLabelRenderer()
                        .renderer(renderer)
                        .args([chartObj.seriesdata, 0]),
                    chartObj.labelHandler,
                    datalabel
                );
            }

            drawTargetMarker(radius, gridList);

            //needle calculation
            var needleWidth = +commonPlotOptions.needle.width / 2 || 7;
            var cir = appendEle(datalabel, 'circle', commonPlotOptions.needle.show ? [1] : []);
            cir.attrs({
                r: needleWidth,
                transform: `translate(${needleWidth},${needleWidth})`
            });
        }

        //calculating new radius
        var arcBound = bbox(dummyArc),
            bb = bbox(container),
            heightPercent = arcBound.height / radius,
            widthPercent = arcBound.width / radius,
            newradius = radius + mathMin((w - bb.width) / widthPercent, (h - bb.height) / heightPercent),
            firedList = {};

        if (isDial) {
            var firedList = fireElementsNgetList();
            var bb = bbox(container);
            newradius = radius + mathMin((w - bb.width) / widthPercent, (h - bb.height) / heightPercent);
            newradius = newradius > maxRadius ? maxRadius : newradius; // Setting maximum to the radius
            newradius = isFinite(newradius) && newradius > 0 ? newradius : 0;
        }

        innerRadius =
            getInnerRadius(commonPlotOptions, newradius) - (orient == INNER_ALIGN ? (+commonPlotOptions.strokeWidth || 0) / 2 : 0);
        axisRadius =
            (orient == INNER_ALIGN
                ? tickOrient == INNER_ALIGN
                    ? innerRadius
                    : innerRadius - (getMinRadius(radius) - radius)
                : newradius) -
            (tickOrient == INNER_ALIGN ? axislineSize / 2 : 0) -
            (orient == INNER_ALIGN ? padding : 0);
        axisComponent.radius(axisRadius);
        if (!firedList.axisgroup) {
            renderAxis();
        }
        arc.outerRadius(newradius).innerRadius(
            !isDial ||
                commonPlotOptions.needle.show ||
                (chartObj.datalabels.moduleExists && isDatalabelEnabled(commonPlotOptions.datalabels, chartObj))
                ? 0
                : innerRadius
        );
        dummyArc.attr('d', arc(1));
        if (isDial) {
            !firedList.minmaxlabels ? drawMinMaxLabel(newradius) : null;
            drawTargetMarker(newradius, gridList);
        }

        //Reference point calculation
        var arcBound = bbox(dummyArc);
        var bb = bbox(container);
        var pt = polarRefPoint(bb, plotarea.width, plotarea.height, newradius, 0);

        // TODO: enhance
        if (chartObj.cache.polarAxis.trimmed) {
            pt[0] = plotarea.width / 2;
        }

        var arcRad =
            (tickOrient == INNER_ALIGN && orient == OUTER_ALIGN
                ? getMinRadius(newradius)
                : newradius - (orient == INNER_ALIGN ? 0 : axislineSize / 2)) - (orient == INNER_ALIGN ? 0 : padding);
        var dwidth = tickOrient == INNER_ALIGN ? (newradius - arcRad) * widthPercent : 0,
            dheight = tickOrient == INNER_ALIGN ? (newradius - arcRad) * heightPercent : 0;

        polarMeta = {
            x: pt[0],
            y: pt[1],
            width: arcBound.width - axislineSize - dwidth,
            height: arcBound.height - axislineSize - dheight,
            radius: arcRad,
            actualRadius: newradius,
            innerRadius,
            axisRadius
        };

        dummyArc.remove();
        datalabel.remove();
        targetMarkerSeries.remove();
        targetMarkerSeries = null;
    }

    _childNodes(container).attr('transform', `translate(${polarMeta.x},${polarMeta.y})`);
    container.style('visibility', null);

    if (chartObj.dataObject.isPolarAxisCategory) {
        container.selectAll(`${PATH_ELEMENT}.domain`).attr('d', () => grid(polarMeta.radius, scaleObj, axis_data));
    }

    chartObj.plotarea_approx = polarMeta;
    bound.polar = polarMeta;

    function getMinRadius(rad) {
        var a = [],
            subtickRad =
                rad + ((orient == OUTER_ALIGN ? axislineSize : axislineSize / 2) + subtickmarkSize * (tickOrient == INNER_ALIGN ? -1 : 1)),
            func = tickOrient == INNER_ALIGN ? d3_min : d3_max;
        wrap.selectAll(TEXT_ELEMENT).each(function(d) {
            var bb = bbox(this),
                textRad =
                    rad +
                    ((orient == OUTER_ALIGN ? axislineSize : axislineSize / 2) + tickmarkSize + ticklabelPad) *
                        (tickOrient == INNER_ALIGN ? -1 : 1);
            var xy = polarToXY(textRad, scale(d));
            var arr = [
                xyToRadius(xy[0] + bb.x, xy[1] + bb.y),
                xyToRadius(xy[0] + bb.x, xy[1] + bb.y + bb.height),
                xyToRadius(xy[0] + bb.x + bb.width, xy[1] + bb.y),
                xyToRadius(xy[0] + bb.x + bb.width, xy[1] + bb.y + bb.height),
                subtickRad
            ];
            a.push(func(arr));
        });
        return func(a);
    }

    function drawMinMaxLabel(radius) {
        let options = { radius, innerRadius, type, index, scale: scaleObj, element: container },
            minmaxLabelSpec = MinmaxLabelSpec(chartObj, options);
        minMaxLabel(minmaxLabelSpec);
    }

    function drawTargetMarker(radius, gridList) {
        // Draw all the target markers. So that we can get correct radius
        gridList.forEach((grid, i) => {
            let wrap = createElement('g', `.targetSeries-${i}`, [1], targetMarkerSeries);
            dialUtils.targetMarker(wrap, radius, innerRadius, chartObj, grid.data, undefined, wrap);
        });
    }

    function fireElementsNgetList() {
        var firedList = {};

        fireElement(firedList, 'axisgroup');
        fireElement(firedList, 'minmaxlabels');
        fireElement(firedList, 'targetSeries');

        return (renderer.firedList = firedList);
    }

    function fireElement(firedList, type) {
        var selector = ['.targetSeries', '.datalabel', '.minmaxlabels', '.axisgroup'].filter(d => d !== `.${type}`).join();
        container.selectAll(selector).style('display', NONE);
        var bb_axis = bbox(container),
            newradius = radius + mathMin((w - bb_axis.width) / widthPercent, (h - bb_axis.height) / heightPercent),
            maxWidth = perToPx(type === 'minmaxlabels' ? commonPlotOptions.minmaxlabels.maxWidth : null, radius, perToPx('50%', radius));

        if (newradius < radius - maxWidth || radius <= 0 || !isFinite(newradius)) {
            firedList[type] = true;
            container.selectAll(`.${type}`).remove();
        }
        container.selectAll(selector).style('display', null);
    }
}

function drawAngularAxisComp(chartObj, axisObj, type = 'y', index = 0, container, bound) {
    var yaxis_data = chartObj.systemConf.chart.axes.yaxis[index],
        show = isAxis(chartObj, type, yaxis_data),
        polarMeta = bound.polar,
        radius = polarMeta.radius;

    const { ticklabel } = yaxis_data,
        ticklabelShowAs = ticklabel.showAs,
        columnInfo = chartObj.dataset.getColumn(type, index);

    var orient = yaxis_data.orient || RIGHT_ALIGN,
        tickSize = validateNullVal(yaxis_data.tickmark.size, 6),
        tickPadding = validateNullVal(yaxis_data.ticklabel[orient == LEFT_ALIGN ? 'marginRight' : 'marginLeft'], 3);

    var yscale = axisObj.scale,
        range = yscale.range(),
        xScale = getAxisObject(chartObj, 'x', 0).scale;

    yscale = yscale.copy().range([range[1], range[0]]); // For tick values, it should be reversed

    var plotarea_approx = {
        left: chartObj.plotarea.left,
        top: chartObj.plotarea.top,
        width: radius,
        height: radius
    };
    var axes = getAxisObject(chartObj, type, index);
    var ticksObj = axis.ticks(chartObj, yaxis_data, 'y', index, orient, plotarea_approx);
    axes.ticks = ticksObj;
    ticksObj.preCalcApproxAxisLength(plotarea_approx);

    // test
    var tickVal = ticksObj.getTickValues(radius, radius, null, null, radius / 20, axes.Ticks);

    var metaData = chartObj.userdata.metadata;
    var isMergedAxis = metaData.axes.y[index].length > 1; //For merged axis, need not to apply formatting
    var columnidx = chartObj.dataset.getColumnIdx(type, index);
    // var datatype = metaData.columns[columnidx].datatype;
    // var formatConf = isMergedAxis && type == 'y' ? merged_axis_column_intersect(index, chartObj) : metaData.columns[columnidx][datatype];
    // var formattingFunction = chartObj.dataObject.isStackedPercentChart && type == 'y' ? getFormattedValueByPercent : getFormattedValue;

    // #ZC2806
    if (isArray(axes.logDomain)) {
        yscale.domain(axes.logDomain);
    }

    var axisObj = axis
        .get()
        .orient(orient)
        .scale(yscale)
        .tickPadding(tickPadding)
        .tickSize(tickSize, 0)
        .plotarea({
            width: radius,
            height: radius
        })
        .radius(radius)
        .type(type)
        .systemconf(chartObj.systemConf)
        .chartObj(chartObj)
        .tickRotation(tickVal.rotation)
        .preCalcTickLen(ticksObj.preCalcTickLen)
        .tickValues(tickVal.ticks)
        .tickFormat(function(d, i) {
            let formattedValue = ticklabelShowAs ? invokeFunction(ticklabelShowAs, d, ticklabel, { data: [columnInfo], index: i }, chartObj) 
                : getFormat(chartObj, type, index, 'axis')(d);
            d3_select(this).text(formattedValue);
        })
        .columnIndices(isMergedAxis && type == 'y' ? metaData.axes.y[index] : [columnidx])
        .polar(true);

    let _transform = getSvgTransform(container).translate;
    container
        .datum(yaxis_data)
        .call(axisObj, type, index)
        .attr('transform', `translate(${polarMeta.x + _transform[0]},${polarMeta.y - radius + _transform[1]})`);

    if (!show) {
        container.remove();
    }

    //! moved to radial-axis parser majorticks. To be removed later in after M2.35
    // var ticks = container.selectAll('.tick');
    // ticks.append(PATH_ELEMENT);
    // var gridLine = ticks.select(PATH_ELEMENT);

    // gridLine
    //     .attr('transform', d => `${'translate(0' + ','}${radius - yscale(d)})`)
    //     .attr('d', d => grid(radius - yscale(d), xScale, yaxis_data))
    //     .styles({
    //         fill: NONE,
    //         stroke: yaxis_data.grid.color,
    //         'stroke-width': yaxis_data.grid.strokeWidth
    //     });
}

function setRange4Scale(chartObj, type) {
    let mode = getMode(chartObj, type),
        facetRendering = isFacet(chartObj);

    if (facetRendering) {
        // Set range
        let axisObjs = getAxisObjects(chartObj, type),
            range;

        axisObjs.forEach((_, i) => {
            let axisConfig = getAxisData(chartObj, type, i);

            if (mode === 'polar') {
                range = getPolarRange(chartObj, type);
            } else {
                let { radius } = chartObj.plotarea_approx;
                range = axisConfig.reversed && !chartObj.dataObject.isstacked() ? [radius, 0] : [0, radius];
            }

            let { scale } = _;
            if (scale) {
                scale.range(range);
                scale.origRange = range;
            }
        });
    } else {
        let axisConfig = getAxisData(chartObj, type, 0);
        // Create scale
        let obj = scale.constructor(chartObj, type == 'x' ? BOTTOM_ALIGN : LEFT_ALIGN);

        // web axis rotated: CX_ZA_SINGLECOL_3_6
        if (chartObj.axes.rotated && type === 'x') {
            chartObj.axes.x.scale = obj[0].scale;
        }

        if (mode !== 'polar') {
            //* update angular scale
            let { radius } = chartObj.plotarea_approx,
                range = axisConfig.reversed && !chartObj.dataObject.isstacked() ? [radius, 0] : [0, radius],
                scale = obj[0].scale;

            scale.range(range);
            scale.origRange = range;
            scale.unknown?.(scale(0));
        }
        // TODO
    }
}

function getMode(chartObj, type) {
    return chartObj.chartTypes[0].value === charttype.dial ? 'polar' : type === 'x' ? 'polar' : 'angular';
}

function isAxis(chartObj, type, axisConfig) {
    let { canvasarea } = chartObj;

    return (
        (axisConfig.show || (axisConfig._backup && axisConfig._backup.show)) &&
        dimension_based_intelligence(chartObj, `${type}axis`, mathMin(canvasarea.width, canvasarea.height))
    );
}

function transformWrap(ele, translate) {
    ele.attr('transform', translateString(translate));
}

function xyToRadius(x, y) {
    return mathSQRT(mathPow(x, 2) + mathPow(y, 2));
}

function grid(r, scaleObj, axisConfig) {
    let range = scaleObj.range();
    if (axisConfig.grid.type === 'arc') {
        var path = d3_arc()
            .innerRadius(r - 1)
            .outerRadius(r)
            .startAngle(range[0])
            .endAngle(range[1])(1);
        return `M${path.split(/L|M/gi)[1]}`;
    }
    var tickValues = scaleObj.ticks ? scaleObj.ticks() : scaleObj.domain();
    var path = tickValues.reduce(function(prev, cur, i) {
        var a = scaleObj(cur);
        var xy = polarToXY(r, a);
        return `${prev}L${xy[0]} ${xy[1]} `;
    }, '');
    return `M${path.substring(1)}Z`;
}

export var _polarAxisRenderer = axis.polar.render;
