// @zohocharts/zc/src/annotation/annotationOptions.js
import { defaultTheme as options } from '../core/defaulttheme';

options.notes = function(userProp) {
    return {
        enabled: false,
        chartValues: [
            //example
            //                      {
            //                        "text": "notesText",
            //                        "x": 0,
            //                        "y": "50%"
            //                      }
        ],
        plotValues: [
            //single series bar example data
            //                              {
            //                                 "data": [2010,57],
            //                                 "text": "High value"
            //                             }
            //multiseries area,stacked area,line,scatter example data
            //                             {
            //                                 "data": [1356978600000,6,"West"],
            //                                 "text": "High value"
            //                             }
        ],
        blockValues: [
            //  {
            //    "htmlEl": "Annotate here",
            //    "hAlign": "center",
            //    "vAlign": "bottom",
            //    "type" : "customNote"
            // }
        ],
        type: 'stickyNote', //stickyNote || customNote
        options: {
            stickyNote: {
                headWidth: 15,
                headHeight: 15,
                bodyWidth: 100,
                bodyHeight: 100,
                fontSize: userProp.fontL1Size.l5,
                color: [
                    '#F0F8FF',
                    '#FFE4C4',
                    '#DCDCDC',
                    '#ADFF2F',
                    '#90EE90',
                    '#87CEFA',
                    '#FFC0CB',
                    '#B0E0E6',
                    '#D2B48C',
                    '#D8BFD8',
                    '#EE82EE',
                    '#F5DEB3'
                ],
                mode: 'close', //open || close
                border: '1px solid grey'
            },
            customNote: {
                fontSize: userProp.fontL1Size.l5
            }
        }
    };
};

export var _annotationOptions = options.notes;
