// @zohocharts/zc/src/annotation/annotation-utils.js
//$Id$
//annotation-utils.js
import { jsonStringify } from '../utils/zcutils';

export function generateDataForAnnotation(d, seriesname) {
    var clonedata = (d.data || d).slice();
    var data = clonedata.splice(0, clonedata.length);
    if (seriesname != null) {
        data.push(seriesname);
    }
    return data;
}

export function saveAnnotation(noteConf, id, text, coordinate, annotationType) {
    if (annotationType == 'plot') {
        noteConf.plotValues.push({
            data: id,
            text: text
        });
    } else {
        noteConf[`${annotationType}Values`].push({
            id: id,
            text: text,
            x: coordinate[0],
            y: coordinate[1]
        });
    }
}

export function getCoordinateWithinBound(left, top, width, height, chartWidth, chartHeight) {
    // TODO I have to check with remila akka.
    if (left + width > chartWidth) {
        left = left - (left + width - chartWidth);
    }
    if (top + height > chartHeight) {
        top = top - (top + height - chartHeight);
    }
    return [left, top];
}

export function getAnnotationIndex(noteConf, id, basedOn) {
    var conf = noteConf[`${basedOn}Values`];
    return conf.indexOf(
        conf.filter(function(d) {
            return basedOn === 'plot' ? jsonStringify(d.data) == jsonStringify(id) : d.id == id;
        })[0]
    );
}
