// @zohocharts/zc/src/annotation/stickyNote.js
//$Id$
//stickyNote.js
export * from './annotationRenderer';
import { annotation } from './annotation';
import { validateNullVal, isArray } from '../utils/zcutils';
import { mathMax, PX } from '../core/zc-constants';
import { d3_select, d3_drag } from '../core/d3-constants';
import { getAnnotationIndex, saveAnnotation, getCoordinateWithinBound } from './annotation-utils';
import { appendEle } from '../utils/zcsvgutils';
import { TRANSPARENT, NONE } from '../core/zc-constants';
import { attachevents } from '../events/eventActions';
import { d3_ease_linear, d3_event } from '../core/d3-constants';
/*
 * z-index check
 * textarea resize
 * While opening a note, check within the plot
 * dragging
 */
annotation.stickyNote = function() {
    var chartObj, conf, coordinate;

    function annotationComp(selection) {
        selection.each(function(args) {
            var container = d3_select(this);
            var noteConf = chartObj.systemConf.notes,
                chartWidth = chartObj.canvasarea.width,
                chartHeight = chartObj.canvasarea.height,
                notesWrap = chartObj.notes;

            var basedOn = args.basedOn || 'chart',
                annotationData = args.data,
                // disabled = args.disabled,
                xy = args.coordinate;

            var xInPercent = typeof xy[0] === 'string' && xy[0].indexOf('%') > -1,
                yInPercent = typeof xy[1] === 'string' && xy[1].indexOf('%') > -1;

            var text = conf.text,
                colorIndex = notesWrap.selectAll('.head').size(),
                bgColor = conf.color,
                open = conf.mode === 'open',
                bodyWidth = validateNullVal(conf.bodyWidth, 100),
                bodyHeight = validateNullVal(conf.bodyHeight, 100),
                headWidth = validateNullVal(conf.headWidth, 15),
                headHeight = validateNullVal(conf.headHeight, 15),
                toggle = open ? 0 : 1,
                cache = basedOn === 'plot' ? args.conf.data : args.conf;
            bgColor = isArray(bgColor) ? bgColor[colorIndex % bgColor.length] : bgColor;

            if (!cache._color) {
                cache._color = bgColor;
            }

            //Drag behaviour
            var drag = d3_drag().on('drag', function(d, i) {
                if (d3_event().defaultPrevented) {
                    return;
                }
                // d3_event().sourceEvent.stopPropagation();
                var parentEle = d3_select(this.parentNode);
                var x = parseFloat(parentEle.style('left')) + d3_event().x,
                    y = parseFloat(parentEle.style('top')) + d3_event().y,
                    w = mathMax(toggle % 2 != 0 ? 0 : bodyWidth, headWidth),
                    h = (toggle % 2 != 0 ? 0 : bodyHeight) + headHeight;
                x = x > chartWidth - w ? chartWidth - w : x < 0 ? 0 : x;
                y = y > chartHeight - h ? chartHeight - h : y < 0 ? 0 : y;
                (coordinate[0] = x), (coordinate[1] = y);
                var leftVal = x + PX,
                    topVal = y + PX;

                parentEle.styles({
                    left: leftVal,
                    top: topVal
                });
                if (basedOn != 'plot') {
                    var iden = parentEle.attr('id');
                    var index = getAnnotationIndex(noteConf, iden, basedOn);
                    if (index != -1) {
                        noteConf.chartValues[index].x = xInPercent ? `${(x / chartWidth) * 100}%` : x;
                        noteConf.chartValues[index].y = yInPercent ? `${(y / chartHeight) * 100}%` : y;
                    }
                }
            });

            container.attr('class', 'group').styles({
                width: 0,
                height: 0,
                'z-index': 2
            });

            // Draw text area
            var textarea = appendEle(notesWrap, 'textarea', [1]);
            textarea
                .styles({
                    width: bodyWidth + PX,
                    height: bodyHeight + PX,
                    position: 'absolute',
                    'border-color': TRANSPARENT,
                    'border-radius': (conf.bodyBorderRadius || 0) + PX,
                    'background-color': TRANSPARENT,
                    overflow: 'hidden',
                    display: NONE,
                    resize: NONE, //Need to implement textarea resize
                    'z-index': 2,
                    outline: NONE
                })
                .on('blur', function(d, i) {
                    var baseEle = textarea.attr('eleID');
                    textarea.style('display', NONE);
                    if (baseEle != null) {
                        var parentEle = d3_select(`#${baseEle}`);
                        var bodyEle = parentEle.selectAll('.body');

                        bodyEle.style('color', null);

                        parentEle.style('z-index', 2);
                    }
                })
                .on('input', function() {
                    var curEle = d3_select(this);
                    if (curEle.style('display') === NONE) {
                        return;
                    }
                    var parentEle = d3_select(`#${curEle.attr('eleID')}`);
                    var noteEle = parentEle.selectAll('.body');
                    var iden = parentEle.attr('id'),
                        txt = textarea.node().value,
                        datum = parentEle.datum();
                    var basedOn = datum.basedOn || 'chart',
                        annotationData = datum.data;
                    noteEle.text(txt);
                    var dt = basedOn == 'plot' ? annotationData : iden;
                    var index = getAnnotationIndex(noteConf, dt, basedOn);
                    if (index == -1) {
                        saveAnnotation(noteConf, dt, txt, xy, basedOn);
                    } else {
                        noteConf[`${basedOn}Values`][index].text = txt;
                    }
                })
                .text(text);

            // Draw head
            var head = appendEle(container, 'div', [1], '.', 'class', 'head');
            head.styles({
                width: headWidth + PX,
                height: headHeight + PX,
                border: conf.border,
                'border-radius': (conf.headBorderRadius || 0) + PX,
                'z-index': 'auto',
                'background-color': cache._color
            });
            // .call(drag);

            var click = function(d, i) {
                if (d3_event().defaultPrevented) {
                    return;
                }
                var close = toggle % 2 == 0;
                var bodyEle = d3_select(this.nextSibling),
                    parentEle = d3_select(this.parentNode);
                textarea.on('blur')();

                if (!close) {
                    var xy = getCoordinateWithinBound(
                        parseFloat(parentEle.style('left')),
                        parseFloat(parentEle.style('top')),
                        bodyWidth,
                        headHeight + bodyHeight,
                        chartWidth,
                        chartHeight
                    );
                    parentEle.styles({
                        left: xy[0] + PX,
                        top: xy[1] + PX
                    });
                } else {
                    parentEle
                        .transition()
                        .delay(250)
                        .duration(0)
                        .styles({
                            left: coordinate[0] + PX,
                            top: coordinate[1] + PX
                        });
                }

                bodyEle
                    .style('color', null)
                    .transition()
                    .ease(d3_ease_linear)
                    .styles({
                        width: close ? 0 + PX : bodyWidth + PX,
                        height: close ? 0 + PX : bodyHeight + PX,
                        display: close ? null : 'block'
                    });

                bodyEle
                    .transition()
                    .delay(250)
                    .style('display', close ? NONE : null);

                notesWrap.selectAll('.group').style('z-index', 1);
                parentEle.style('z-index', 2);

                toggle++;
            };
            var mouseHandlers = {
                    click: click
                },
                touchHandlers = {
                    tap: click
                };
            attachevents(head, mouseHandlers, touchHandlers);

            //Draw body
            var body = appendEle(container, 'div', [1], '.', 'class', 'body');
            body.styles({
                width: !open ? '0' : bodyWidth + PX,
                height: !open ? '0' : bodyHeight + PX,
                display: !open ? NONE : null,
                border: conf.border,
                overflow: 'hidden',
                'box-shadow': conf.shadow,
                'border-radius': (conf.bodyBorderRadius || 0) + PX,
                'z-index': 'auto',
                'background-color': cache._color
            }).text(text);

            var click = function(d, i) {
                var curEle = d3_select(this),
                    parentEle = d3_select(this.parentNode);
                textarea.on('blur')();
                textarea
                    .styles({
                        display: null,
                        left: parentEle.style('left'),
                        top: parseFloat(parentEle.style('top')) + headHeight + PX,
                        color: parentEle.style('color'),
                        'font-size': parentEle.style('font-size'),
                        'font-family': parentEle.style('font-family'),
                        'font-weight': parentEle.style('font-weight'),
                        'font-style': parentEle.style('font-style')
                    })
                    .attr('eleID', parentEle.attr('id'));
                textarea.node().value = curEle.text();
                textarea.node().focus();

                curEle.style('color', TRANSPARENT);

                notesWrap.selectAll('.group').style('z-index', 1);
                parentEle.style('z-index', 2);
                textarea.style('z-index', 3);
            };
            var mouseHandlers = {
                    click: click
                },
                touchHandlers = {
                    tap: click
                };
            attachevents(body, mouseHandlers, touchHandlers);

            let customHTML = appendEle(container, 'div', conf.htmlEl ? [1] : [], '.', 'class', 'zcannotation-custom');
            customHTML
                .styles({
                    position: 'absolute',
                    left: 0,
                    top: 0
                })
                .each(function() {
                    let comp = annotation
                        .customNote()
                        .conf(conf)
                        .chartObj(chartObj);
                    d3_select(this).call(comp);
                });
        });
    }

    annotationComp.chartObj = function(_) {
        if (!arguments.length) {
            return chartObj;
        }
        chartObj = _;
        return annotationComp;
    };
    annotationComp.conf = function(_) {
        if (!arguments.length) {
            return conf;
        }
        conf = _;
        return annotationComp;
    };
    annotationComp.coordinate = function(_) {
        if (!arguments.length) {
            return coordinate;
        }
        coordinate = _;
        return annotationComp;
    };
    return annotationComp;
};

export var _stickyAnnotation = annotation.stickyNote;
