// @zohocharts/zc/src/mark/extender/markSelection.js
import { MarksRegistry } from '../../plot/MarksRegistry.js';
import { select } from './select.js';
import { selectAll } from './selectAll.js';
import { remove } from './remove.js';
import { removeAll } from './removeAll.js';
import { append } from './append.js';
import { insert } from './insert.js';
import { each } from './each.js';
import { attr, attrs } from './attr.js';
import { style, styles } from './style.js';
import { data } from './data.js';
import { get } from './get.js';
import { set } from './set.js';
import { config } from './config.js';
import { simpleClone } from '../../utils/zcutils.js';

export function markSelection(parser) {
    this.parser = { ...parser };
}

markSelection.prototype.select = select;
markSelection.prototype.selectAll = selectAll;
markSelection.prototype.remove = remove;
markSelection.prototype.removeAll = removeAll;
markSelection.prototype.append = append;
markSelection.prototype.insert = insert;
markSelection.prototype.each = each;
markSelection.prototype.attr = attr;
markSelection.prototype.attrs = attrs;
markSelection.prototype.style = style;
markSelection.prototype.styles = styles;
markSelection.prototype.data = data;
markSelection.prototype.get = get;
markSelection.prototype.set = set;
markSelection.prototype.config = config;

MarksRegistry.add('markSelection', markSelection);
