// @zohocharts/zc/src/legend/helpers/select-all/fade-on-select-all.js
import { d3_namespaces } from '../../../core/d3-constants';
import { NULL } from '../../../core/zc-constants';
import { isObjectPropDefined, defined, pick } from '../../../utils/zcutils';
import { createElement } from '../../../utils/zcsvgutils';
import { isFacet } from '../../../axis/hierarchy-axis-utils';
import { getGridList } from '../../../core/data-processor/pivot-layout-helpers';
import { effects } from '../../../effects/zceffects';
import { elementPriority } from '../../legend-helper';
import { getPlotSize } from '../../../plot/plotSize';

export function fadeOnSelectAll(enable, chartObj, opacity, forceEnable, order) {
    let hasMultipleColorLegend = isObjectPropDefined(chartObj.dataObject, 'hierarchical.pivotLayout.layers.groups.length', true) > 1,
        _enable = isEnabled(chartObj, forceEnable, order),
        _color = 'white';

    if (isFacet(chartObj) && hasMultipleColorLegend) {
        let overLaySVG = createElement('svg', '.zc-fade-overlay-svg', [1], chartObj.container.selectAll('.canvas-component')),
            overLayClipHolder = createElement('g', '.zc-fade-overlay-clip-holder', [1], overLaySVG),
            overLay = createElement('rect', '.zc-fade-overlay', [1], overLayClipHolder);

        overLaySVG
            .attrs({
                version: '1.1',
                xmlns: d3_namespaces.svg
            })
            .styles({
                position: 'absolute',
                left: 0,
                top: 0,
                width: chartObj.getChartDimension('width'),
                height: chartObj.getChartDimension('height'),
                'box-sizing': 'border-box',
                'pointer-events': 'none'
            });

        if (!_enable) {
            fadeAll(false, chartObj, opacity);
            fadeAxis(chartObj, false);
            overLay.style('visibility', 'hidden');
            return;
        }
        fadeAll(true, chartObj, opacity);

        let clipPathID,
            desc = '',
            gridList = getGridList(chartObj.dataObject),
            count = 0;

        gridList.forEach(grid => {
            let { bound, yaxiscolumnorder } = chartObj.rendererConf[count].params,
                { row, layer, data } = grid,
                state = chartObj.legend.state[`clr-${layer}`],
                allSelected = defined(forceEnable) && `clr-${layer}` === `clr-${order}` ? forceEnable : state.allSelected;

            fadeAxis(chartObj, !allSelected, opacity, `y-axis-${row}`);
            desc += !allSelected ? getGridBound(chartObj, bound) : '';
            count += data.length;
        });

        var properties = { d: desc, name: 'zc-legend-selectAll', id: chartObj.id };
        clipPathID = desc ? effects.get(effects.svg_clips, 'polygon', properties, { chartObj }) : null;

        overLay
            .attrs({
                x: chartObj.plotarea.left,
                y: chartObj.plotarea.top,
                width: getPlotSize(chartObj, 'width'),
                height: getPlotSize(chartObj, 'height'),
                transform: `translate(${-chartObj.scrolledX || 0},${-chartObj.scrolledY || 0})`,
                'clip-path': clipPathID ? `url(${chartObj.defsLocation}#${clipPathID})` : null
            })
            .styles({
                fill: _color,
                opacity: 1 - opacity,
                visibility: clipPathID ? null : 'hidden'
            });
    } else {
        fadeAll(_enable, chartObj, opacity);
    }
}

function fadeAll(enable, chartObj, opacity) {
    let ele = chartObj.container.selectAll('g.chartgroup,svg.static,canvas.zc-canvas,canvas.zc-labelholder-canvas,div.zcnotes');
    ele.style('opacity', enable ? NULL : opacity);
    elementPriority(ele, enable ? NULL : ['opacity']);
}

function isEnabled(chartObj, forceEnable, order) {
    if (isFacet(chartObj)) {
        let _enable = false;
        if(chartObj.legends){
            chartObj.legends.forEach(legend => {
                if (legend.meta.encode === 'clr') {
                    let state = (chartObj.legend.state && chartObj.legend.state[legend.id]) || { allSelected: true };
                    _enable =
                        _enable ||
                        (defined(forceEnable) && legend.id === `clr-${order}` ? forceEnable : legend.enabled ? state.allSelected : true);
                }
            });
        }

        return _enable;
    } else {
        let state = chartObj.legend.states || {};
        return pick(forceEnable, state.allSelected);
    }
}

function fadeAxis(chartObj, fade, opacity, id) {
    let axisWrap;
    if (id) {
        axisWrap = chartObj.container.selectAll(`g.${id}`);
    } else {
        axisWrap = chartObj.container.selectAll('g.y-axis').selectAll(function() {
            return this.childNodes;
        });
    }

    axisWrap.style('opacity', fade ? opacity : null);
}

function getGridBound(chartObj, bound) {
    let { x, y, width, height } = bound,
        { left: plotarea_left, top: plotarea_top } = chartObj.plotarea;

    return `M${plotarea_left + x} ${plotarea_top + y} L${plotarea_left + x + width} ${plotarea_top + y} L${plotarea_left +
        x +
        width} ${plotarea_top + y + height} L${plotarea_left + x + width} ${plotarea_top + y + height} L${plotarea_left +
        x} ${plotarea_top + y + height} Z`;
}
