// @zohocharts/zc/src/legend/background/legend-background-svg-parser.js
//$Id$
import { TRANSPARENT, NONE, NULL } from '../../core/zc-constants';
import { Mark } from '../../mark/mark-utils';
import { Registry } from '../../core/zc-registry';
import { invokeFunction, getBorderRadius, parseShortHandValue } from '../../utils/zcutils';

// TODO: remove
function LegendBackgroundSvgParser(spec) {
    let config = spec.config,
        backgroundSpec = parseBackground(config.background),
        borderSpec = parseBorder(config.border),
        margin = parseShortHandValue(spec.margin || config.padding),
        backgroundBound;

    return {
        mark: '_rect',
        selector: '.zc-legend-background',
        order: ':first-child', // TODO: remove by preRender
        config: {
            ...backgroundSpec,
            ...borderSpec,
            // TODO: Shadow properties
            x: () => backgroundBound.x, // TODO: work on here
            y: () => backgroundBound.y,
            width: () => backgroundBound.w,
            height: () => backgroundBound.h
        },
        onStart: ele => {
            ele.style('display', NONE);
            invokeFunction(spec.onStart, ele);
            backgroundBound = polyBound(spec.bound(), margin, borderSpec.radius); // TODO: bound(id) like vega
        },
        onEnd: ele => {
            ele.style('display', NULL);
            invokeFunction(spec.onEnd, ele);
        }
    };
}

function _LegendBackgroundSvgParser(spec) {
    let config = spec.config,
        backgroundSpec = parseBackground(config.background),
        borderSpec = parseBorder(config.border),
        margin = parseShortHandValue(spec.margin || config.padding);

    let backgroundMark = Mark('_rect', '.zc-legend-background');
    backgroundMark.dependency = 'legend-content';
    backgroundMark.onStart = spec.onStart;
    backgroundMark.onEnd = spec.onEnd;
    backgroundMark.config = {
        ...backgroundSpec,
        ...borderSpec,
        // TODO: Shadow properties
        x: asParser('legend-content', 'background_x', -margin.left),
        y: asParser('legend-content', 'background_y', -margin.top),
        width: asParser('legend-content', 'background_width', margin.left + margin.right),
        height: asParser('legend-content', 'background_height', margin.top + margin.bottom)
    };

    return backgroundMark;
}

function asParser(name, as, margin) {
    return { _type: 'as', name, as, onEnd: _ => _ + margin };
}

function parseBackground(config) {
    let { color = TRANSPARENT, alpha } = config;

    return {
        fillColor: color,
        fillOpacity: alpha
    };
}

function parseBorder(config) {
    let { show, color, size, alpha, style, lineCap, radius } = config;

    if (show) {
        return {
            strokeColor: color,
            strokeWidth: size,
            strokeOpacity: alpha,
            dashStyle: style,
            lineCap,
            radius
        };
    }

    return { strokeWidth: 0, radius: 0 };
}

function polyBound(bb, margin, radius) {
    return {
        x: bb.x - margin.left,
        y: bb.y - margin.top,
        w: bb.width + margin.left + margin.right,
        h: bb.height + margin.top + margin.bottom,
        r: getBorderRadius(radius)
    };
}

export { LegendBackgroundSvgParser, _LegendBackgroundSvgParser };
Registry.setComponent('legendBackgroundSvg.parser', _LegendBackgroundSvgParser);
