// @zohocharts/zc/src/sunburst/sunburstchart.js
//$Id$
//sunburstchart.js
import { Registry } from '../core/zc-registry';
import { svg_plot_renderer, HIGHLIGHT_EFFECT_REGEX, labelRenderer } from '../core/main-zc-init';
import { utils, validateNullVal, defined, isArray, perToPx } from '../utils/zcutils';
import { POINTER, NONE, NULL } from '../core/zc-constants';
import { SUNBURST_CHARTNAME, PIE_CHARTNAME } from '../core/chartname-constants';
import { getAnimationConf, cancelAnimation } from '../utils/animationutils';
import { mathMin, mathMax, mathAbs, mathPI, TRANSPARENT } from '../core/zc-constants';
import { getFillColor, getStrokeColor, getColorCopy, getRawColor, getColorByIndex } from '../utils/colorutils';
import {
    d3_arc,
    d3_ease_linear,
    d3_select,
    d3_interpolate,
    d3_partition,
    d3_ascending,
    d3_range,
    d3_scaleLinear,
    d3_event,
    d3_hierarchy
} from '../core/d3-constants';
import { FAN_ANIMATION, WHEEL_ANIMATION } from '../core/animation-constants';
import { setPointerEvents } from '../events/events-utils';
import { attachevents } from '../events/eventActions';
import { PATH_ELEMENT } from '../core/element-constants';
import { appendEle } from '../utils/zcsvgutils';
import { isDatalabelEnabled } from '../label/datalabelsUtils';
import { isBoundBoxEnabled, getBoundBoxPadding, getBoundBoxStrokeWidth } from '../label/boundBoxUtils';
import { PERCENT_SHOWAS } from '../label/showas-constants';
import { getEventConf, getEventObject } from '../events/events-utils';
import { generateDataForAnnotation } from '../annotation/annotation-utils';
import { getChildrenCount } from '../core/hierarachy-data-processor';
import { updateSelectionOnDatapoint } from '../events/event-update-selection';
import { isFacet } from '../axis/hierarchy-axis-utils';
import { setOpacityForTickFilterElements, setPivotAttributes } from '../plot/svg/set-pivot-attr';

let annotation = Registry.getComponent('annotation');
/**
 *
 * @returns {chart} sunburst renderer.
 */
/*
 * PENDING:
 * Annotation for parent arcs. Legend filter
 * Annotation for hierarchy data
 */
svg_plot_renderer.sunburst = function() {
    var commonRendererProp,
        xscale,
        yscale,
        arc,
        radius,
        dataset,
        redraw = false,
        arcLabel,
        systemconf,
        oldData,
        yaxiscolumnorder;
    var chartwidth, chartheight, color, chartObj, notesArray, path, pathGroup, commonPlotOptions;
    var start,
        end,
        innerrad,
        rad,
        highlightGroup,
        highlightEle,
        highlightHierarchy,
        maxDepth,
        valueFunction,
        nodes,
        stratifiedIndices = {};

    function chart(selection) {
        selection.each(function(data) {
            //renderer config values
            var rp = commonRendererProp.params;
            (chartObj = rp.chartObj),
                (commonPlotOptions = rp.commonPlotOptions),
                (color = rp.color),
                (notesArray = rp.notesArray),
                (yaxiscolumnorder = rp.yaxiscolumnorder),
                (dataset = rp.dataset);
            systemconf = chartObj.systemConf;
            var { labelQueue, bound, rendererIndex } = rp;
            let colorScheme = commonPlotOptions.colorScheme || 'tints'; // solid || tints || shades

            (chartwidth = bound.width), (chartheight = bound.height);
            let paneid = isFacet(chartObj) ? `${data[0].row}_${data[0].column}` : null;
            var gvpFn = utils.getValueOfPlotOptions,
                dataLabels = commonPlotOptions.datalabels,
                outerPadding = +commonPlotOptions.outerPadding || 0,
                NegativeAsAbsolute = commonPlotOptions.showAsNegative === 'absolute',
                finalDisplay = d => {
                    var hidden = isDisabled(data, d) || start(d) === end(d) || !d.value; // if disabled or value = 0 | null
                    return hidden ? NONE : NULL;
                },
                initialDisplay = function(d) {
                    var prevDisplay = d3_select(this).style('display'),
                        curDisplay = finalDisplay(d);
                    return !redraw || (prevDisplay === NONE && curDisplay === NONE) ? curDisplay : NULL;
                },
                padding;
            highlightGroup = null;
            maxDepth = 0;
            nodes = [];

            var showlabel = chartObj.datalabels.moduleExists && isDatalabelEnabled(dataLabels, chartObj);
            var boundBoxPadding =
                showlabel && isBoundBoxEnabled(dataLabels, chartObj)
                    ? getBoundBoxPadding(dataLabels, chartObj) + getBoundBoxStrokeWidth(dataLabels, chartObj)
                    : 0;
            highlightHierarchy = commonPlotOptions.events && commonPlotOptions.events.highlightHierarchy;

            //Animation variables
            var animationOptions = getAnimationConf(commonPlotOptions, null, chartObj, SUNBURST_CHARTNAME);
            var isAnimationEnable = animationOptions.enabled,
                animationType = animationOptions.type,
                easingType = animationOptions.easingType,
                transitionDuration = redraw ? (isAnimationEnable ? 200 : 0) : animationOptions.duration,
                transitionEnd = function() {
                    d3_select(this).style('display', finalDisplay);
                };

            radius = mathMin(chartwidth, chartheight) / 2 - outerPadding - 2 * boundBoxPadding;
            var innerRadius = mathMin(mathMax(0, perToPx(commonPlotOptions.innerRadius, radius)), perToPx('95%', radius)),
                fillColor = [],
                strokeColor = [];

            (start = function(d) {
                return mathMax(0, mathMin(2 * mathPI, xscale(d.x0)));
            }),
                (end = function(d) {
                    return mathMax(0, mathMin(2 * mathPI, xscale(d.x1)));
                }),
                (innerrad = function(d) {
                    return yscale(d.y0) + padding / 2;
                });
            rad = function(scaleObj, extra) {
                extra = extra || 0;
                return function(d) {
                    return yscale(d.y1) + extra - padding / 2;
                };
            };
            valueFunction = function(d) {
                if (isDisabled(data, d)) {
                    return 0;
                }
                var y = dataset.getY(d, yaxiscolumnorder) || 0;
                if (y < 0) {
                    return NegativeAsAbsolute ? mathAbs(y) : 0;
                }
                return y;
            };

            arc = d3_arc() //#ZC814 - d3.svg.arc has issue in this
                .startAngle(start)
                .endAngle(end)
                .innerRadius(innerrad)
                .outerRadius(rad(yscale));

            arcLabel = d3_arc() //#ZC814 - d3.svg.arc has issue in this
                .startAngle(start)
                .endAngle(end)
                .innerRadius(innerrad)
                .outerRadius(rad(yscale));

            pathGroup = appendEle(selection, 'g', [data], '.', 'class', 'hierarchychart');
            pathGroup.attr('transform', `translate(${chartwidth / 2},${chartheight / 2})`);

            oldData = [];
            pathGroup.selectAll(PATH_ELEMENT).each(function(d) {
                oldData.push({
                    x0: d.x0,
                    x1: d.x1
                });
            });
            var packedData = getPartitionNodeDescendants(data);
            chart.total = packedData[0].value;
            path = appendEle(pathGroup, PATH_ELEMENT, packedData);
            cancelAnimation(path);

            var colorLighter = [],
                leafNodes = [],
                seriesWiseNodes = [];
            data.forEach(function(d, i) {
                fillColor[i] = chart.fillColorFunc = getFillColor(chartObj, i, paneid, [radius, radius]);
                strokeColor[i] = getStrokeColor(chartObj, i, paneid);

                leafNodes[i] = [[]];
                leafNodes[i].disabled = d.disabled;
                leafNodes[i].plotoptions = d.plotoptions;
                seriesWiseNodes[i] = [];
                seriesWiseNodes[i].disabled = d.disabled;
                seriesWiseNodes[i].plotoptions = d.plotoptions;
                var nonRGBAColor = getColorByIndex(chartObj, i, paneid);
                colorLighter[i] = d3_scaleLinear()
                    .domain([0, getChildrenCount(d)])
                    .range([
                        nonRGBAColor, 
                        getColorCopy(nonRGBAColor, colorScheme == 'solid' ? 0 : colorScheme == 'shades' ? -1.5 : 1.5, 'brighter')
                    ]);
            });

            var plotLevelOptions = d => d.depth && seriesWiseNodes[getSeriesIndex(d)].plotoptions,
                isEventEnabled = d => {
                    var plotOptions = plotLevelOptions(d);
                    return validateNullVal(getEventConf(plotOptions, commonPlotOptions, chartObj, 'enabled'), true);
                },
                cursor = d => {
                    var plotOptions = plotLevelOptions(d);
                    return isEventEnabled(d) ? getEventConf(plotOptions, commonPlotOptions, chartObj, 'cursor') || POINTER : NULL;
                };

            path.attr('class', function(d) {
                return getLevel(d) != null ? `zc_sunburst series_${getSeriesIndex(d)}` : 'zc_sunburst';
            })
                .attr('id', function(d) {
                    return `arcPath_${getLevel(d)}`;
                })
                .each(function(d, i) {
                    d.index = i;
                    var seriesIndex = getSeriesIndex(d),
                        level = getLevel(d);
                    maxDepth = mathMax(maxDepth, d.depth);
                    d.data._seriesIndex = seriesIndex;
                    d.data.paneid = paneid;
                    if (!d.depth) {
                        d.data.arcColor = TRANSPARENT;
                    } else {
                        if (d.data.color) {
                            d.data.arcColor = d.data.color;
                        } else {
                            var plotoptions = data[seriesIndex].plotoptions;
                            var chClr =
                                plotoptions != null
                                    ? gvpFn(plotoptions, commonPlotOptions)('childrenColors')
                                    : commonPlotOptions.childrenColors;
                            if (!d.children && chClr != null) {
                                d.data.arcColor = chClr[stratifiedIndices[level] % chClr.length];
                            } else {
                                var c = colorLighter[seriesIndex](stratifiedIndices[level]);
                                d.data.arcColor = c;
                            }
                        }

                        if (!d.children) {
                            leafNodes[seriesIndex][0].push(d);
                        }

                        seriesWiseNodes[seriesIndex].push(d);
                        nodes[level] = d;
                    }
                });

            xscale = d3_scaleLinear().range([0, 2 * mathPI]);
            yscale = d3_scaleLinear()
                .range([innerRadius, radius])
                .domain([packedData[0].y1, 1]);
            padding = mathMin(mathMax(0, perToPx(commonPlotOptions.padding, radius, 0)), ((radius - innerRadius) * 0.9) / maxDepth);

            path.style('display', initialDisplay)
                .transition()
                .duration(transitionDuration)
                .ease(redraw ? d3_ease_linear : easingType)
                .attrTween(
                    'd',
                    redraw || animationType === FAN_ANIMATION ? fanTween : animationType === WHEEL_ANIMATION ? wheelTween : innerTween
                )
                .on('end', transitionEnd);

            path.styles({
                fill: function(d) {
                    let seriesIndex = getSeriesIndex(d);

                    return (d.fillColor = fillColor[seriesIndex || 0](seriesIndex, d.data));
                },
                stroke: function(d) {
                    let seriesIndex = getSeriesIndex(d);

                    if (!d.depth) {
                        return 'white';
                    } else if (commonPlotOptions.strokeColor) {
                        return commonPlotOptions.strokeColor;
                    } else {
                        return strokeColor[seriesIndex || 0](seriesIndex, d.data);
                    }
                },
                'stroke-opacity': function(d) {
                    return d.depth ? gvpFn(data[getSeriesIndex(d)].plotoptions, commonPlotOptions)('strokeOpacity') : 1;
                },
                'fill-opacity': function(d) {
                    return d.depth ? gvpFn(data[getSeriesIndex(d)].plotoptions, commonPlotOptions)('fillOpacity') : 0;
                },
                'stroke-width': function(d) {
                    return d.depth ? gvpFn(data[getSeriesIndex(d)].plotoptions, commonPlotOptions)('strokeWidth') : 0;
                },
                cursor: cursor,
                'pointer-events': d => (d.depth && isEventEnabled(d) ? null : 'none')
            });

            setPivotAttributes(path, chartObj, 'sunburst');
            setOpacityForTickFilterElements(path, chartObj, 'sunburst');

            setPointerEvents(chartObj, path, d => !isEventEnabled(d), paneid);
            if (!chartObj.cache.search) {
                chartObj.cache.search = {};
            }
            var mousemove = function(e, d) {
                    var args = getEventArgs(e, d);
                    chartObj.eventHandler.handlers.mousemove.call(chartObj.eventHandler, e, args);
                    updateSelectionOnDatapoint(d, args);
                    chartObj.cache.search.lastPane = paneid;
                },
                mouseout = function(e, d) {
                    var args = getEventArgs(e, d);
                    chartObj.eventHandler.handlers.mouseout.call(chartObj.eventHandler, e, args);
                    updateSelectionOnDatapoint(d, args);
                    chartObj.cache.search.lastPane = null;
                },
                click = function(e, d) {
                    var args = getEventArgs(e, d);
                    chartObj.eventHandler.handlers.click.call(chartObj.eventHandler, e, args);
                    updateSelectionOnDatapoint(d, args);
                    chartObj.cache.search.lastPane = paneid;
                    //clickFunc(path,d);     //default click function
                },
                dblclick = function(e, d) {
                    if (!chart.chartRendered) {
                        return;
                    }
                    var dt =
                        d.children == null
                            ? !annotation._empty
                                ? generateDataForAnnotation(d, data.length > 1 ? d.parent.data.seriesname : null)
                                : d
                            : [d.data.seriesname, d.value];
                    var coordinates = arcLabel.centroid(d);
                    coordinates = [coordinates[0] + chartwidth / 2, coordinates[1] + chartheight / 2];
                    var args = getEventArgs(e, d);
                    args.coordinate = coordinates;
                    args.basedOn = 'plot';
                    args.data = dt;
                    chartObj.eventHandler.handlers.doubleclick.call(chartObj.eventHandler, e, args, getSeriesIndex(d), d);
                    updateSelectionOnDatapoint(d, args);
                    chartObj.cache.search.lastPane = paneid;
                },
                eventAttacherOptions = {
                    target: 'plot',
                    touchleave: true,
                    chartObj: chartObj
                };
            var mouseHandlers = {
                    mousemove: mousemove,
                    mouseout: mouseout,
                    click: click,
                    dblclick: dblclick
                },
                touchHandlers = {
                    singletouchmove: mousemove,
                    touchend: mouseout,
                    tap: click,
                    dbltap: dblclick
                };
            attachevents(
                path.filter(d => d.depth && isEventEnabled(d)),
                mouseHandlers,
                touchHandlers,
                eventAttacherOptions
            );

            chartObj.timerObj[chartObj.timerObj.length] = chart.renderedTimer = setTimeout(function() {
                if (labelRenderer != null) {
                    var outline = d3_arc()
                        .startAngle(start)
                        .endAngle(end)
                        .innerRadius(rad(yscale))
                        .outerRadius(rad(yscale));

                    var datalabelsPadding = commonPlotOptions.datalabels.padding + boundBoxPadding;
                    var outlineWithPad = d3_arc()
                        .startAngle(start)
                        .endAngle(end)
                        .innerRadius(rad(yscale, datalabelsPadding))
                        .outerRadius(rad(yscale, datalabelsPadding));

                    data.forEach(function(d, i) {
                        let _currentRenderer = chartObj.renderer[isFacet(chartObj) ? rendererIndex + i : 0],
                            _currentRendererIndex = isFacet(chartObj) ? rendererIndex + i : i,
                            _paneid = isFacet(chartObj) ? `${d.row}_${d.column}` : i;

                        var getArguments = function(nodes, isParent) {
                            return [
                                leafNodes,
                                i,
                                nodes,
                                radius,
                                function(d, showAs) {
                                    if (showAs === PERCENT_SHOWAS) {
                                        return chart.total;
                                    } else {
                                        var node = chart.getProcessedNode(d);
                                        return node.parent.value;
                                    }
                                },
                                outlineWithPad,
                                outline,
                                arcLabel,
                                [chartwidth / 2, chartheight / 2],
                                isParent
                            ];
                        };

                        labelQueue.push({
                            renderer: _currentRenderer,
                            chartObj: chartObj,
                            charttype: PIE_CHARTNAME,
                            rendererIndex: _currentRendererIndex,
                            paneid: _paneid,
                            args: getArguments(seriesWiseNodes[i])
                        });
                        labelQueue.push({
                            renderer: _currentRenderer,
                            chartObj: chartObj,
                            charttype: 'pieInnerLabels',
                            rendererIndex: _currentRendererIndex,
                            paneid: _paneid,
                            args: getArguments(leafNodes[i][0])
                        });
                        labelQueue.push({
                            renderer: _currentRenderer,
                            chartObj: chartObj,
                            charttype: 'pieInnerLabels',
                            rendererIndex: _currentRendererIndex,
                            paneid: _paneid,
                            args: getArguments(
                                seriesWiseNodes[i].filter(function(d) {
                                    return d.children;
                                }),
                                true
                            )
                        });
                    });
                }

                //Annotation preprocessing
                if (!annotation._empty) {
                    data.forEach(function(series, i) {
                        notesArray = annotation.preprocessAnnotations(
                            data,
                            i,
                            function(dt, k, j) {
                                var node = chart.getProcessedNode(dt);
                                var c = arcLabel.centroid(node);
                                return [c[0] + chartwidth / 2, c[1] + chartheight / 2];
                            },
                            notesArray,
                            chartObj,
                            rp
                        );
                    });
                }

                redraw = true;
                chart.chartRendered = true;
            }, transitionDuration + 1000 / 60);

            function getEventArgs(e, d) {
                var ev = getEventObject(e),
                    level = getLevel(d);
                return {
                    renderer: chart,
                    seriesIndex: getSeriesIndex(d),
                    // "subseriesPos": 0,
                    // "itemPos": d.itemPos,
                    paneid: paneid,
                    rendererIndex: rendererIndex,
                    coordinates: [ev.pageX, ev.pageY],
                    element: this,
                    data: d.data,
                    level: level,
                    selected: d.selected,
                    _key: `ZC_${chartObj.id}_${level}`
                };
            }
        });
        return chart;
    }

    chart.commonRendererProp = function(_) {
        if (!arguments.length) {
            return commonRendererProp;
        }
        commonRendererProp = _;
        return chart;
    };
    chart.defaultMouseOverFunc = function(d, seriesPos, itemPos, paneid) {
        createHighlightElement();
        var highlightEffect = chartObj.legendHighlight.getLegendHighlightEffect(seriesPos, 'selectedSeries', 'plot', paneid),
            elementIds = getHierarchyElementIds(getLevel(d));

        if (highlightEffect !== NONE) {
            highlightGroup.style('opacity', null);

            let zchighlighterPosition = [commonRendererProp.params.bound.x, commonRendererProp.params.bound.y];

            var refElements = pathGroup.selectAll(elementIds.join()).nodes();
            highlightEle.each(function(d, i) {
                var ele = d3_select(this);
                if (refElements[i]) {
                    var refEle = d3_select(refElements[i]),
                        arcColor = getRawColor(chartObj, seriesPos, refEle.datum(), seriesPos, paneid);
                    ele.attr('d', refEle.attr('d')).styles(
                        chartObj.legendHighlight.getEffect(
                            highlightEffect,
                            SUNBURST_CHARTNAME,
                            HIGHLIGHT_EFFECT_REGEX.test(highlightEffect) ? arcColor : TRANSPARENT,
                            { fill: TRANSPARENT, stroke: arcColor },
                            seriesPos,
                            'plot'
                        )
                    );
                    ele.attr('transform', `translate(${zchighlighterPosition[0]},${zchighlighterPosition[1]})`);
                } else {
                    ele.attr('d', null);
                }
            });
        }
    };
    chart.defaultMouseOutFunc = function(d, i) {
        createHighlightElement(); //Touch device -> tap on a data, then tap a legend item. Mouseout should be called and legend tap action also will get executed.

        highlightGroup.style('opacity', 0);
    };
    chart.getProcessedNode = function(d) {
        return nodes[getLevel(d)];
    };

    // function Tween(d, tweentype) {
    //     var xd = d3_interpolate(xscale.domain(), [d.x, d.x + d.dx]),
    //         yd = d3_interpolate(yscale.domain(), [d.y, 1]),
    //         yr = d3_interpolate(yscale.range(), [d.y ? 20 : 0, radius]);
    //     if (tweentype == "arc") {
    //         return function(d, i) {
    //             return i ?

    //                 function(t) {
    //                     return arcLabel(d);
    //                 } :
    //                 function(t) {
    //                     xscale.domain(xd(t));
    //                     yscale.domain(yd(t)).range(yr(t));
    //                     return arcLabel(d);
    //                 };
    //         };
    //     } else {
    //         return function(d, i) {
    //             return i ?

    //                 function(t) {
    //                     return "translate(" + arcLabel.centroid(d) + ")";
    //                 } :
    //                 function(t) {
    //                     xscale.domain(xd(t));
    //                     yscale.domain(yd(t)).range(yr(t));
    //                     return "translate(" + arcLabel.centroid(d) + ")";
    //                 };
    //         };
    //     }
    // }

    function createHighlightElement() {
        if (highlightGroup) {
            return;
        }

        var g = chartObj.staticPaneSVG.selectAll('g#zchierarchyhighlight');
        if (!g.empty()) {
            //#ZC1756
            highlightGroup = g;
            highlightGroup.attr(
                'transform',
                `translate(${chartObj.plotarea.left + chartwidth / 2},${chartObj.plotarea.top + chartheight / 2})`
            );
            highlightEle = appendEle(highlightGroup, PATH_ELEMENT, d3_range(highlightHierarchy ? maxDepth : 1));
        }
    }

    function fanTween(d, i) {
        var startAngle = d3_interpolate(redraw && oldData[i] ? start(oldData[i]) : 0, start(d));
        var endAngle = d3_interpolate(redraw && oldData[i] ? end(oldData[i]) : 0, end(d));
        return function(t) {
            arc.startAngle(startAngle(t)).endAngle(endAngle(t));
            return arc(d);
        };
    }

    function innerTween(d) {
        var inner = d3_interpolate(0, innerrad(d));
        var outer = d3_interpolate(0, rad(yscale)(d));
        return function(t) {
            arc.innerRadius(inner(t)).outerRadius(outer(t));
            return arc(d);
        };
    }

    function wheelTween(d) {
        var endAngle = d3_interpolate(start(d), end(d));
        return function(t) {
            arc.endAngle(endAngle(t));
            return arc(d);
        };
    }

    // function clickFunc(path, d) {
    //     path.transition()
    //         .duration(250)
    //         .attrTween("d", Tween(d, "arc"));
    //     // text.style("visibility", function(e) {return isParentOf(d, e) ? null : "hidden";})
    //     //  .transition()
    //     //  .duration(250)
    //     //  .attrTween("transform",Tween(d,"label"))
    //     //  .style("fill-opacity", function(e) { return isParentOf(d, e) ? 1 : 1e-6; });
    // }

    // function getDomainMax(parentElement) {
    //     return parentElement.children.length;
    // }

    // function isParentOf(p, c) {
    //     if (p === c) {
    //         return true;
    //     }
    //     if (p.children) {
    //         return p.children.some(function(d) {
    //             return isParentOf(d, c);
    //         });
    //     }
    //     return false;
    // }

    function getSeriesIndex(d) {
        var level = getLevel(d);
        if (level) {
            return +level.split('_')[0];
        }
    }

    function getHierarchyElementIds(id) {
        var hierarchyIds = [];
        var currentId = `${PATH_ELEMENT}#arcPath`;

        if (highlightHierarchy) {
            var levels = id.split('_');

            levels.forEach(function(d) {
                currentId = `${currentId}_${d}`;
                hierarchyIds.push(currentId);
            });
        } else {
            hierarchyIds.push(`${currentId}_${id}`);
        }

        return hierarchyIds;
    }

    function getLevel(d) {
        return d.level || (d.data && d.data.level);
    }

    function isDisabled(data, d) {
        var seriesIndex = getSeriesIndex(d);
        return defined(seriesIndex) && data[seriesIndex].disabled;
    }

    var childrenCount;
    function getChildren(node) {
        var level = getLevel(node),
            depth = level && level.split('_').length;
        stratifiedIndices[level] = depth === 1 ? (childrenCount = 0) : childrenCount;

        if (!chartObj.dataObject.hasHierarchyData && depth == 1 && node.length < 2 && node[0] && isArray(node[0][0])) {
            return node[0];
        } else if (isArray(node[0])) {
            return node;
        } else {
            childrenCount++;
            return null;
        }
    }

    function getPartitionNodeDescendants(data) {
        var partition = d3_partition();
        var root = d3_hierarchy(data, getChildren)
            .sum(valueFunction)
            .sort(d3_ascending);

        partition(root);
        return root.descendants();
    }

    return chart;
};

export var _sunburstChart = svg_plot_renderer.sunburst;
