// @zohocharts/zc/src/scale/operators/extremes.js
import { d3_extent, d3_merge } from '../../core/d3-constants';
import { invokeFunction, isArray } from '../../utils/zcutils';
import { isFunction } from '../../utils/function-utils';
import { ALLOWED_EXTREMES_OPERATORS } from './allowed-operators';
import { ScaleRegistry } from '../scale-registry';

export function getExtremes(operators) {
    let _scale = this,
        _chart = _scale.chartObj,
        _scaleType = _scale.type,
        _encodeType = _scale.encode,
        _reversed = _scale.reverse,
        _dimension = _scale.dimension,
        // _field = _scale.field,
        // _dataIndex = _field.dataIndex,
        _extremes;

    _extremes = operators.map(operator => {
        if (isArray(operator)) {
            return operator;
        }

        if (isFunction(operator)) {
            return invokeFunction(operator, this);
        }

        let appliedOperators = {};

        if (ALLOWED_EXTREMES_OPERATORS.has(operator._type)) {
            appliedOperators[operator._type] = ALLOWED_EXTREMES_OPERATORS.get(operator._type).call(this, operator);
            return appliedOperators[operator._type];
        }
    });
    let value = d3_extent(d3_merge(_extremes));
    _scale.config.tempDomain = d3_extent(d3_merge(_extremes));
    _scale.config.extremes = value;
    return value;
}
export function getSeriesExtremes(operator) {
    let _scale = this,
        _scaleType = _scale.type,
        _encodeType = _scale.encode,
        _reversed = _scale.reverse,
        _dimension = _scale.dimension,
        _encodeId = _scale.id,
        key = _scale.key,
        seriesExtermes = operator.seriesExtermes;

    let values = d3_merge(seriesExtermes[key]);
    let extremes = values.length != 0 ? d3_extent(values) : values.enabled ? [0, 0] : null; // yvalues.enabled for referring that axis is visible or not.If the axis is visible,but all the data contains 0 or null,we dont want to remove axis
    return extremes;
}

//x data encode and stack layout indices
export function getExtremesByAccessor(operator) {
    let _scale = this,
        _chart = _scale.chartObj || {},
        _scaleType = _scale.type,
        _encodeType = _scale.encode,
        _reversed = _scale.reverse,
        _dimension = _scale.dimension,
        _getAccessor = operator.accessor, //mandatory
        _data = operator.data,
        _indices = operator.indices,
        { condition } = operator;

    let extremesByAccessor;

    extremesByAccessor = _data.map(function(d, i) {
        let extremes;
        var data = d3_merge(d.data),
            seriesIdx = _indices ? _indices[i] : i,
            _accessor = _getAccessor(_encodeType, seriesIdx);

        if (condition) {
            data = data.filter(condition);
        }

        if (operator.isRangeType && operator.isRangeType(seriesIdx)) {
            var mergedYranges = d3_merge(data.map(_accessor, _encodeType, seriesIdx));
            extremes = d3_extent(mergedYranges);
        } else {
            extremes = d3_extent(data, _accessor, _encodeType, seriesIdx);
        }
        return extremes;
    });

    return d3_extent(d3_merge(extremesByAccessor));
}

//levelmarker
//extremesByDataIndices
export function getExtremesByData(operator) {
    let _scale = this,
        _chart = _scale.chartObj,
        _accessor = operator.accessor,
        _data = operator.data || _chart.seriesdata || this.config.data,
        _dataIndices = operator.indices,
        _extents = [],
        extremes;

    _data.forEach(function(d) {
        if (!d.disabled) {
            // var subSeries = d3_merge(d.data || d);
            var subSeries = d.data || d;
            eachData(subSeries, function(point) {
                _dataIndices.map(d => {
                    _extents = _extents.concat(point[d]);
                });
            });

            // subSeries.forEach(function(point) {
            //     _dataIndices.map(d => {
            //         _extents = _extents.concat(point[d]);
            //     });
            // });
        }
    });

    extremes = d3_extent(_extents);
    return extremes;
}

function eachData(data, callback) {
    if (data && isArray(data) && data.every(_ => isArray(_))) {
        data.forEach(function(child) {
            eachData(child, callback);
        });
    } else {
        callback(data);
    }
}
export function getLogDomainExtremes(operators) {
    let _scale = this,
        _extremes;

    _extremes = operators.map(operator => {
        if (isArray(operator)) {
            return operator;
        }
    });
    let value = d3_extent(d3_merge(_extremes));
    _scale.config.extremes = value;
    return value;
}

ScaleRegistry.add('axis.extremes', getExtremes);
