// @zohocharts/zc/src/legend/events/visual-effects/size-nested/highlight-item.js
import { ORDINAL_DATATYPE } from '../../../../core/datatype-constants';
import { NestedView } from '../../../../view/nested/nested-view';
import { mergeJSON, isEmptyObject, defined } from '../../../../utils/zcutils';
import { getHoverOptions } from '../../legend-event-helpers';

export function HighlightItem(mouse, data) {
    let manager = this,
        { options } = manager,
        { config, highlightElement, viewScale, viewBound, dataType } = options,
        { ranges, nested: nestedConfig } = config,
        hoverOptions = getHoverOptions(config),
        { marker: markerConfig } = nestedConfig,
        { marker: hoverMarkerConfig } = hoverOptions,
        point = data.point;

    if (isEmptyObject(hoverMarkerConfig)) {
        return;
    }

    // Show highlight element
    highlightElement.style('opacity', null);

    let viewParser = {
        bound: viewBound,
        data: getHighlightData(point, ranges, dataType),
        scale: viewScale,
        config: mergeJSON(hoverMarkerConfig, markerConfig)
    };
    let viewObj = new NestedView(viewParser);
    viewObj.create();
    viewObj.draw(highlightElement);
}

function getHighlightData(point, ranges, dataType) {
    if (dataType === ORDINAL_DATATYPE) {
        return [ranges[point][0]];
    } else if (ranges && ranges.length) {
        return [ranges[point]];
    } else if (defined(point.data)) {
        return [point.data];
    } else {
        return [];
    }
}
