// @zohocharts/zc/src/axis/cartesian/parser/marks/cartesian-subminorticks.js
import { ParserRegistry } from "../../../../core/parser-registry";
import { isArray, pick } from "../../../../utils/zcutils";

export function _cartesianSubMinorTicks(data) {
    const self = this;

    const { component, config, helpers } = this,
        { subtickmark } = config,
        { position: positionH } = helpers,
        {subtickmark: _position } = positionH;

    const _x_tm = positionH.tickmark.attrs_key.x;

    const subMinorTicksMarks = {
        mark: 'g',
        selector: '.zc-subMinorTicks',
        data: component.isSubTickMarkShow && isArray(component.minorGrid) ? [1] : [],
        children: [
            {
                name: 'sub-minor-ticks',
                selector: '.zc-sub-ticks',
                mark: 'line',
                data: component.isSubTickMarkShow ? component.minorGrid : [],
                config: {
                    strokeWidth: pick(subtickmark.strokeWidth, 1)
                },
                properties: {
                    attrs: {
                        [`${_x_tm}1`]: _position?.[`${_x_tm}1`],
                        [`${_x_tm}2`]: _position?.[`${_x_tm}2`],
                        transform: function (d) {
                            return positionH.tickG.transform(d);
                        }
                    },
                    styles: {
                        fill: 'none',
                        'shape-rendering': 'crispedges',
                        stroke: subtickmark.color
                    }
                }
            }
        ]
    };
    return subMinorTicksMarks;
}
ParserRegistry.axis('cartesian.subminorticks', _cartesianSubMinorTicks);
