// @zohocharts/zc/src/axis/cartesian/parser/marks/cartesian-domain-marks.js
import { ParserRegistry } from "../../../../core/parser-registry";
import { pick } from "../../../../utils/zcutils";

const KEY = 'axisline';
export function _cartesianDomain(data) {
    const self = this;

    const axisline = this.config.axisline,
        {show, strokeWidth, style: dashStyle, lineCap, color} = axisline,
        helpers = this.helpers,
        positionH = helpers?.position || {},
        _position = positionH[KEY] || {};

    const domainMarks = {
        mark: 'path',
        selector: '.domain',
        data: [null],
        order: '.tick',
        config: {
            strokeWidth: pick(strokeWidth, 1),
            dashStyle: dashStyle,
            lineCap: lineCap
        },
        properties: {
            attrs: {
                d: function() {
                    return _position.path || positionH.axislinePath;
                }
            },
            styles: {
                stroke: show === false ? 'transparent' : color || '#000',
                fill: 'none'
            }
        }
    };
    return domainMarks;
}
ParserRegistry.axis('cartesian.domain', _cartesianDomain);
