// @zohocharts/zc/src/axis/cartesian/parser/marks/cartesian-ticks-marks.js
import { d3_interpolate, d3_map, d3_select } from "../../../../core/d3-constants";
import { ParserRegistry } from "../../../../core/parser-registry";
import { addSVGtitleEle, createElement } from "../../../../utils/zcsvgutils";
import { pick, validateNullVal } from "../../../../utils/zcutils";
import { updateMinmaxlabelPosition } from "../../../minmaxLabelPosition";

export function _cartesianTicks(data) {
    const self = this;

    const { chart, helpers, component } = this,
        { plotarea, axes } = chart,
        { position: positionH, transition: transition_ } = helpers,
        racing = chart.racing?.syncTransition;

    // updating tickvalues
    const tickValues = component.tickValues == null
        ? component.scale.ticks
            ? component.scale.ticks.apply(component.scale, [])
            : component.scale.domain()
        : component.tickValues;

    if (this.axisTicks?.tickInfo) {  // handled for radial ticks
        this.axisTicks.tickInfo.ticklabelObj = tickValues;
    }
    const tickMap = d3_map();
    tickValues.map((d, i) => tickMap.set(d, i));

    const ticklabel_onEnd = ticklabelOnEnd.call(this, { tickValues, tickMap });
    const attrs = {};

    if (!component.syncTransition) {
        attrs.transform = function (d) {
            return positionH.tickG.transform(d);
        }
    }

    // add ticks children marks here
    const minorTicksMarks = minorTicks.call(this, data);
    const majorTicksMarks = majorTicks.call(this, data);
    const tickLabelMarks = tickLabel.call(this, data, { tickValues, tickMap });
    const ticksChildrenMarks = [minorTicksMarks, majorTicksMarks, tickLabelMarks];

    const ticksMarks = {
        mark: 'g',
        selector: '.tick',
        data: tickValues,
        dataAccessor: function (d) {
            if (chart.racing) {
                d3_select(this).interrupt()
            }
            return component.scale(d);
        },
        // on: {
        //     start: function (ele) {
        //         if (chart.racing?.interrupt) {
        //             ele.interrupt();
        //         }
        //     }
        // },
        config: {
            enter:
                racing && self.type === 'x'
                    ? {
                        opacity: 1e-6,
                        transform: function () {
                            var pos;
                            if (axes.rotated && axes.y.groups[0].prerenderComp.orient === 'bottom') {
                                pos = `0,-20`;
                            } else {
                                pos = `0,${plotarea.height + 20}`;
                            }
                            if (!self.chart.racing.keyFramedRace) {
                                pos = `${plotarea.width},${0}`;
                            }
                            return `translate(${pos})`;
                        }
                    }
                    : {
                        opacity: 1e-6,
                        transform: function (d) {
                            return positionH.tickG?.transform(d);
                        }
                    }
        },
        anime: {
            update: {
                transition: transition_,
                opacity: 1,
                transform: racing
                    ? {
                        property: 'attrTween',
                        to: raceTick_positionUpdate(positionH, self, ticklabel_onEnd)
                    }
                    : function (d) {
                        let _position = positionH.tickG?.position(d);
                        this.__axis = _position;
                        return positionH.tickG?.transform(d, _position);
                    }
            },
            exit: racing && self.type === 'x'
                ? {
                    transition: transition_,
                    opacity: {
                        property: "styleTween",
                        to: function () {
                            var i = d3_interpolate(1, 1e-6)
                            return function (t) {
                                return i(t * 3);
                            }
                        }
                    },
                    transform: function () {
                        var pos;
                        if (axes.rotated && axes.y.groups[0].prerenderComp.orient === 'bottom') {
                            pos = `0,-20`;
                        } else {
                            pos = `0,${plotarea.height + 20}`;
                        }
                        return `translate(${pos})`;
                    }
                }
                : {
                    transition: transition_,
                    opacity: 1e-6,
                    transform: function (d) {
                        return isFinite(positionH.tickG?.position(d))
                            ? positionH.tickG?.transform(d)
                            : this.getAttribute('transform');
                    }
                }
        },
        properties: {
            attrs,
            props: {
                __index__: function (d, i) {
                    return i;
                }
            }
        },
        children: ticksChildrenMarks
    }
    return ticksMarks;
}
ParserRegistry.axis('cartesian.ticks', _cartesianTicks);

function minorTicks(data) {
    const { strokeWidth, color } = this.config.tickmark,
        positionH = this.helpers.position,
        {tickmark: _position} = positionH,
        _x_tm = positionH.tickmark?.attrs_key?.x;

    const minorTicksMarks = {
        name: 'tickmark',
        selector: '.minorTicks',
        mark: 'line',
        data: function (d, i) {
            return [d];
        },
        config: {
            strokeWidth: pick(strokeWidth, 1)
        },
        properties: {
            attrs: {
                [`${_x_tm}1`]: _position?.[`${_x_tm}1`],
                [`${_x_tm}2`]: _position?.[`${_x_tm}2`]
            },
            styles: {
                fill: 'none',
                'shape-rendering': 'crispedges',
                stroke: color
            }
        }
    };
    return minorTicksMarks;
}

function majorTicks(data) {
    const KEY = 'gridline';
    const { component, config, helpers } = this,
        { position: positionH } = helpers,
        _position = positionH[KEY] || {},
        { grid } = config;

    const majorTicksMarks = {
        name: 'gridline',
        selector: '.majorTicks',
        mark: 'line',
        data: function (d, i) {
            return component.isGridShow && component.renderGrid && positionH.gridline ? [d] : [];
        },
        config: {
            strokeWidth: pick(grid.strokeWidth, 1),
            dashStyle: grid.style,
            lineCap: grid.lineCap
        },
        properties: {
            attrs: {
                x1: _position.x1,
                x2: _position.x2,
                y1: _position.y1,
                y2: _position.y2
            },
            styles: {
                fill: 'none',
                'shape-rendering': 'crispedges',
                stroke: grid.color
            }
        }
    };
    return majorTicksMarks;
}

function tickLabel(data, options) {
    const KEY = 'ticklabel';
    const self = this;
    const { chart, component, helpers, config } = this,
        { ticklabel } = config,
        { position: positionH, styles } = helpers,
        styleObj = styles[KEY] || styles,
        { ticklabel: ticklabelH = {} } = positionH;

    const racing = chart.racing?.syncTransition,
        updating = chart.updating;

    const _x_tl = ticklabelH.attrs_key?.x, //tl -ticklabel, tm- tickmark
        _y_tl = ticklabelH.attrs_key?.y;

    const events_tick = ticklabel.events,
        eventDisabled = events_tick != null && validateNullVal(events_tick.enabled, true) == false,
        cursor = eventDisabled ? 'auto' : events_tick?.cursor || 'pointer';

    // updating tickValues
    const { tickValues, tickMap } = options;

    const ticklabel_onEnd = ticklabelOnEnd.call(this, options);

    const tickLabelMarks = {
        name: 'ticklabel',
        selector: '.ticklabel',
        mark: 'text',
        datum: () => { },
        data: d => (component.renderTicklabel ? [d] : []),
        dataAccessor: function (d) {
            return component.scale(d);
        },
        on: {
            end: racing && updating ? null : ticklabel_onEnd
        },
        config: {
            enter: {}
        },
        properties: {
            attrs: {
                [_x_tl]: d => {
                    let val = positionH.ticklabel?.[_x_tl];
                    return positionH.ticklabel?.customTransform?.(d, _x_tl, val);
                },
                [_y_tl]: d => {
                    let val = positionH.ticklabel && positionH.ticklabel[_y_tl],
                        customTransform = positionH.ticklabel && positionH.ticklabel.customTransform;
                    return customTransform && customTransform(d, _y_tl, val);
                },
                dy: d => {
                    return self.axisTicks?.tickInfo?.mode == 'zigzag' && tickMap.get(d) % 2
                        ? positionH.ticklabel?.dy
                        : positionH.ticklabel?._dy;
                },
                transform: d => positionH.ticklabel?.transform1?.(d, tickMap.get(d))
            },
            styles: {
                'text-anchor': ticklabelH.textAnchor,
                cursor: cursor,
                ...styleObj
            }
        }
    };
    return tickLabelMarks;
}

function raceTick_positionUpdate(positionH, axisConf, ticklabel_onEnd) {
    var _ = axisConf;
    return function (d, i) {
        let _position = positionH.tickG.position(d);
        var from_transform = this.getAttribute('transform'),
            to_transform = positionH.tickG.transform(d, _position),
            transform = d3_interpolate(from_transform, to_transform),
            tickLabels = d3_select(this.parentElement)
                .selectAll('.tick')
                .selectAll('.ticklabel');
        return function (t) {
            ticklabel_onEnd(tickLabels);
            if (t == 1 && _.chart.updating) {
                var _timeout = setTimeout(function () {
                    ticklabel_onEnd(tickLabels);
                    clearTimeout(_timeout);
                }, 1);
                //formating gets called before the animation of the ticks gets ended.
            }
            return transform(t);
        };
    }
}

function ticklabelOnEnd(options) {
    const self = this;
    const { chart, component, helpers, renderer, config, commonOption } = this,
        orient = this.orient || commonOption.orient,
        container = typeof this.container == 'function' ? this.container() : this.container,
        positionH = helpers.position;

    // updating tickValues
    const { tickValues, tickMap } = options;

    const ticklabel_onEnd = (d, i) => {
        //format
        let f = component.tickFormat == null
            ? component.scale.tickFormat
                ? component.scale.tickFormat.apply(component.scale, [])
                : d => d
            : component.tickFormat,
            //update ticklabel element by data, so we can add text only to to the active text node
            t = createElement('g', '.tick', tickValues, container, null, d => component.scale(d), {
                remove: false
            }),
            t1 = t.select('text.ticklabel');

        updateMinmaxlabelPosition(t1, orient, positionH.ticklabel, config.ticklabel, config, component.scale, chart);

        t1.each(function (d, i) {
            f.call(this, this.__data__, tickMap?.get(d));
        });
        if (config.ticklabel.tooltip === true) {
            d.each(function (d) {
                addSVGtitleEle(d3_select(this), this.formattedVal);
            });
        }
    }
    return ticklabel_onEnd;
}
