// @zohocharts/zc/src/axis/cartesian/parser/marks/cartesian-axislabel-marks.js
import { ParserRegistry } from "../../../../core/parser-registry";

export function _cartesianAxisLabel (data) {
    const self = this;

    const { helpers, container, axisTitle } = this,
        { title: axisTitleH } = helpers;

    const axisLabelMarks = {
        mark: 'g',
        selector: '.axislabelgroup',
        parent: container.node().parentElement,
        data: [1],
        children: [
            {
                mark: 'text',
                selector: '.axislabel',
                data: axisTitle.hideTitle ? [] : [axisTitleH.text],
                properties: {
                    text: function(d, i) {
                        return d;
                    },
                    attrs: {
                        x: axisTitleH.x,
                        y: axisTitleH.y,
                        dy: axisTitleH.dy,
                        transform: axisTitleH.transform
                    },
                    styles: {
                        'text-anchor': axisTitleH.textAnchor,
                        ...axisTitleH.styleObj
                    }
                }
            }
        ]
    };
    return axisLabelMarks;
}
ParserRegistry.axis('cartesian.axislabel', _cartesianAxisLabel);
