// @zohocharts/zc/src/axis/cartesian/parser/marks/cartesian-submajorticks.js
import { ParserRegistry } from "../../../../core/parser-registry";
import { isArray, pick } from "../../../../utils/zcutils";

export function _cartesianSubMajorTicks(data) {
    const self = this;

    const { component, config, helpers } = this,
        { minorGrid } = config,
        { position: positionH } = helpers,
        { gridline } = positionH;

    const subMajorTicksMarks = {
        mark: 'g',
        selector: '.zc-subMajorTicks',
        data: component.isMinorGridShow && isArray(component.minorGrid) ? [1] : [],
        children: [
            {
                name: 'sub-major-ticks',
                selector: '.zc-sub-major-ticks',
                mark: 'line',
                data: positionH.gridline && component.isMinorGridShow ? component.minorGrid : [],
                config: {
                    strokeWidth: pick(minorGrid.strokeWidth, 1),
                    dashStyle: minorGrid.style,
                    lineCap: minorGrid.lineCap
                },
                properties: {
                    attrs: {
                        x1: gridline?.x1,
                        x2: gridline?.x2,
                        y1: gridline?.y1,
                        y2: gridline?.y2,
                        transform: function (d) {
                            return positionH.tickG.transform(d);
                        }
                    },
                    styles: {
                        fill: 'none',
                        'shape-rendering': 'crispedges',
                        stroke: minorGrid.color
                    }
                }
            }
        ]
    };
    return subMajorTicksMarks;
}
ParserRegistry.axis('cartesian.submajorticks', _cartesianSubMajorTicks);
