// @zohocharts/zc/src/behaviours/listeners/zoom/zoom-d3.js
import { boundToRect, each, isArray, mergeJSON, pick } from '../../../utils/zcutils';
import { KEYBOARD_EVENT_HANDLER } from '../../../zoom/keybord-event-handler';
import { d3_dispatch, d3_event, d3_zoom, d3_zoomIdentity, d3_zoomTransform } from '../../../core/d3-constants';
import { mathMax, mathMin } from '../../../core/zc-constants';
import { Registry } from '../../../core/zc-registry';

/**
 *
 * @param options
 * @constructor
 */

const defaultOptions = {
    // * For wheel zoom
    wheelZoom: true,
    // * for touch zoom feature
    doubleClickZoom: false,
    // * Panning behaviour
    pan: true,
    // * min and max zoom
    scaleExtent: [1, 8],
    extent: [[0, 0], [0, 0]],
    translateExtent: [[0, 0], [0, 0]],
    // meta key for mouse zoom
    metaKey: null,

    transitionDuration: 750
};

/**
 * This is used as a interface for zoom position
 */
export class ZoomD3Options {
    constructor(options) {
        this.settings = { ...defaultOptions, ...options };
    }
}

export class ZoomD3 {
    /**
     *
     * @param {ZoomD3Options} zoomD3Options
     */
    constructor(zoomD3Options) {
        this.options = zoomD3Options.settings;
        this.initialize();
        this.hookCount = 0;
    }

    initialize() {
        let zoom = this,
            options = zoom.options;

        zoom._maxZoom = mathMax(...options.scaleExtent);
        zoom._minZoom = mathMin(...options.scaleExtent);
        /**
         *
         * @type {ZoomBehavior}
         */
        zoom.d3_zoom = d3_zoom().scaleExtent([zoom._minZoom, zoom._maxZoom]);

        // create the signals for callback
        zoom._signal = d3_dispatch('start', 'zoom', 'end');

        zoom.enableSignals();

        if (options.filter) {
            zoom.d3_zoom.filter(options.filter);
        }

        if (options.element && options.element.isD3Selection) {
            zoom.bindTo(options.element);
        }
    }

    /**
     *
     */
    enableSignals() {
        var zoom = this,
            d3_zoom = zoom.d3_zoom,
            signal = zoom._signal;

        d3_zoom.on('start', (event) => {
            if (zoom.__notify !== false) {
                signal.call('start', zoom, d3_event(event));
            }
        });

        d3_zoom.on('zoom', (event) => {
            if (zoom.__notify !== false) {
                signal.call('zoom', zoom, d3_event(event));
            }
        });

        d3_zoom.on('end', (event) => {
            if (zoom.__notify !== false) {
                signal.call('end', zoom, d3_event(event));
            }
            zoom.__notify = true;
        });
    }

    /**
     *
     * @return {{}}
     */
    getSignals() {
        return this._signal;
    }

    /**
     *
     * @param type
     * @param hook
     * @return {string}
     */
    addHook(type, hook) {
        this.hookCount++;
        let currentHookID = `${type}.${this.hookCount}`;
        this._signal.on(currentHookID, hook);
        return currentHookID;
    }

    /**
     *
     * @param currentHookID
     */
    removeHook(currentHookID) {
        this._signal.on(currentHookID, null);
    }

    /**
     *
     * @param {HTMLElement} element
     */
    bindTo(element) {
        let zoom = this,
            d3_zoom = this.d3_zoom,
            options = this.options;

        zoom._element = element;
        zoom._elementNode = element.node();

        // set the extent
        d3_zoom.extent(zoom.__getExtent());
        d3_zoom.translateExtent(zoom.__getTranslateExtent());

        if (options.metaKey) {
            KEYBOARD_EVENT_HANDLER.ADD_HOOK(options.metaKey, function(active) {
                if (active) {
                    element.call(d3_zoom);
                    zoom.constrainEvent();
                } else {
                    element.on('.zoom', null);
                }
            });
        } else {
            element.call(d3_zoom);
            zoom.constrainEvent();
        }
    }

    // should be called only after binding the event.
    constrainEvent() {
        let options = this.options,
            element = this._element;

        if (options.doubleClickZoom === false) {
            element.on('dblclick.zoom', null);
        }

        if (options.wheelZoom === false) {
            element.on('wheel.zoom', null);
        }

        if (options.pan === false) {
            element.on('mousedown.zoom', null);
        }

        if (options.touchZoom === false) {
            element.on('touchstart.zoom', null);
        }
    }

    /**
     *
     * @return {d3.event.transform|*}
     */
    getTransform() {
        return (d3_event() && d3_event().transform) || (this._element && d3_zoomTransform(this._element.node())) || d3_zoomIdentity;
    }

    step(stepSize, transitionDuration) {
        stepSize = stepSize <= 0 ? 1 / -stepSize : stepSize;
        this.d3_zoom.scaleBy(this.__getZoomTransition(transitionDuration), stepSize);
    }

    /**
     *
     * @param scale
     * @param bounds
     */
    zoomToBounds(bounds, scale, transitionDuration) {
        this.zoomToRect(boundToRect(bounds), scale, transitionDuration);
    }

    /**
     *
     * @param scale
     * @param rect
     */
    zoomToRect(rect, scale, transitionDuration) {
        this.zoomToPoint(ZoomD3.centroid(rect), pick(scale, this.__getConstrainedScale(rect)), transitionDuration);
    }

    /**
     *
     * @param point
     * @param scale
     * @param transitionDuration
     */
    zoomToPoint(point, scale, transitionDuration) {
        let zoom = this;

        scale = zoom.__limitScale(pick(scale, zoom._maxZoom));
        point = zoom.__constrainPoint(point, scale);
        zoom.interpolateZoom(point, scale, transitionDuration);
    }

    /**
     *
     * @param translate
     * @param scale
     * @param transitionDuration
     */
    interpolateZoom(translate, scale, transitionDuration) {
        this.zoomToIdentity(ZoomD3.getIdentity(translate, scale), transitionDuration);
    }

    zoomToIdentity(identity, transitionDuration, notify = true) {
        let zoom = this;
        zoom.__notify = notify;
        zoom.__getZoomTransition(transitionDuration).call(zoom.d3_zoom.transform, identity);
    }

    /**
     * reset the zoom
     */
    reset() {
        this._element
            .transition()
            .duration(this.options.transitionDuration)
            .call(this.d3_zoom.transform, d3_zoomIdentity);
    }

    /**
     *
     * @param xy
     * @return {*}
     */
    invert(xy) {
        return this.getTransform().invert(xy);
    }

    update(option) {
        let zoom = this,
            options = (zoom.options = mergeJSON(zoom.options, option, true, true));

        zoom.d3_zoom.scaleExtent(options.scaleExtent).translateExtent(zoom.__getTranslateExtent());
    }

    updateExtent(extent) {
        if (isArray(extent) && isArray(extent[0]) && isArray(extent[1])) {
            this.options.extent = extent;
            this.d3_zoom.extent(extent);

            this.options.translateExtent = extent;
            this.d3_zoom.translateExtent(extent);

            if (this._element) {
                this._element.call(this.d3_zoom);
                // after call we should constrain the event
                this.constrainEvent();
            }
        }
    }

    /**
     *
     * @param translate
     * @param scale
     * @return {Transform}
     */
    static getIdentity(translate, scale) {
        return d3_zoomIdentity.translate(translate[0], translate[1]).scale(scale);
    }

    /**
     *
     * @return {[[number, number], [number, number]]}
     * @private
     */
    __getExtent() {
        return this.options.extent || [[0, 0], [this._elementNode.clientWidth, this._elementNode.clientHeight]];
    }

    destroy() {
        let zoom = this;
        zoom._signal = null;
        // #ZC2766
        zoom._element && zoom._element.on('.zoom', null);
        each(zoom, function(value, key) {
            zoom[key] = null;
            delete zoom[key];
        });
    }

    /**
     *
     * @return {[[number, number], [number, number]]}
     * @private
     */
    __getTranslateExtent() {
        return this.options.translateExtent || this.__getExtent();
    }

    /**
     *
     * @return {Transition}
     * @private
     */
    __getZoomTransition(transitionDuration) {
        return this._element.transition().duration(pick(transitionDuration, this.options.transitionDuration));
    }

    /**
     *
     * @param boundRect
     * @return {number}
     * @private
     */
    __getConstrainedScale(boundRect) {
        let zoom = this,
            translateExtentRect = boundToRect(zoom.__getTranslateExtent());

        return mathMax(
            1,
            mathMin(zoom._maxZoom, 1 / mathMax(boundRect.width / translateExtentRect.width, boundRect.height / translateExtentRect.height))
        );
    }

    /**
     *
     * @param point
     * @param scale
     * @return {number[]}
     * @private
     */
    __constrainPoint(point, scale) {
        let translateExtentRect = boundToRect(this.__getTranslateExtent());
        return [
            mathMin(
                translateExtentRect.x,
                mathMax(translateExtentRect.width / 2 - scale * point[0], translateExtentRect.width - translateExtentRect.width * scale)
            ),
            mathMin(
                translateExtentRect.y,
                mathMax(translateExtentRect.height / 2 - scale * point[1], translateExtentRect.height - translateExtentRect.height * scale)
            )
        ];
    }

    __limitScale(scale) {
        return mathMax(this._minZoom, mathMin(scale, this._maxZoom));
    }

    /**
     *
     * @param {{x: number, y: number, width: number, height: number}} rect
     * @return {number[]}
     */
    static centroid(rect) {
        return [+rect.x + +rect.width / 2, +rect.y + +rect.height / 2];
    }
}

ZoomD3.prototype.on = ZoomD3.prototype.addHook;

Registry.setComponent('ZoomD3', ZoomD3);
Registry.setComponent('ZoomD3Options', ZoomD3Options);
