// @zohocharts/zc/src/core/data-getter/chartSpec_new.js
import { DataSplicer } from './chartSpec_standard';
import { Registry } from '../zc-registry';
import { getAxisLayout } from '../../axis/axisutils';
import { pick, mergeJSON, defined } from '../../utils/zcutils';

export class DataSplicer4ViewParser extends DataSplicer {
    constructor(data, systemConf, chart) {
        super(data, systemConf, chart);
        this.axisConfig = {
            x: new Array(0),
            y: new Array(0)
        };
    }
    getColumns() {
        if (this.data.columns) {
            return this.data.columns;
        }
        return super.getColumns();
    }
    getEncode() {
        if (this.data.encoding) {
            return this.data.encoding;
        }
        return super.getEncode();
    }
    getAxisConfig(type, order) {
        order = pick(order, 0);
        let layout = getAxisLayout(null, type, this.chart.axes.rotated);
        if (this.chart.viewFlow) {
            let axisView = this.chart.viewFlow.axes[layout],
                axisConfig_merged = this.axisConfig[type][order];
            //if new parser does not initialize axis parser or axis spec
            if (!axisView) {
                return super.getAxisConfig(type, order);
            }
            if (!defined(order)) {
                return axisView;
            }
            if (!axisView[order]) {
                return super.getAxisConfig(type, order);
            }
            if (axisConfig_merged) {
                return axisConfig_merged;
            } else {
                let oldConfig = super.getAxisConfig(type, order);
                if(type == 'y' && order > 0){
                    if(oldConfig.merged){
                        this.axisConfig[type][order] = mergeJSON(axisView[order].conf.spec, oldConfig, false, false);
                    }else{
                        return {};//dummy
                    }
                }
                this.axisConfig[type][order] = mergeJSON(axisView[order].conf.spec, oldConfig, false, false);
            }
            return this.axisConfig[type][order];
        }
    }
    getPlotConfig(data, index) {
        return data.spec;
    }
    rotated(){
        return this.rotated;
    }
    getPropValue(key, value, from) {}
}

Registry.setComponent('dataSplicer4ViewParser', DataSplicer4ViewParser);
