// @zohocharts/zc/src/axis/axisOptions.js
import { defaultTheme as options } from '../core/defaulttheme';

options.axes = function(userProp) {
    var axisOptions = {
        xaxisLevels: {},
        yaxisLevels: {},
        xaxis: {
            splitLeaves: false,
            uniformLeafScale: true,
            show: true,
            label: {
                show: true,
                text: 'Xaxis',
                fontSize: userProp.fontL2Size.l4,
                marginTop: 5,
                marginRight: 5,
                marginBottom: 5,
                marginLeft: 5
            },
            grid: {
                show: true,
                color: 'transparent',
                strokeWidth: null
            },
            minorGrid: {
                show: false,
                color: '#ececec'
            },
            subtickmark: {
                show: false,
                color: '#ececec',
                size: 3,
                count: 0
            },
            ticklabel: {
                textOverflow: 'ellipsis', // none,ellipsis,hidden
                fontSize: userProp.fontL2Size.l5
            },
            tickmark: {
                strokeWidth: null,
                color: '#ececec',
                size: 3
            },
            axisline: {
                show: true,
                color: '#ececec',
                strokeWidth: null,
                style: 'solid'
            },
            baseline: {
                show: false,
                value: 10,
                color: 'black'
            },
            threshold: {
                line: {
                    show: false,
                    values: [],
                    label: {
                        text: ['Description comes here'],
                        fontSize: userProp.fontL2Size.l5
                    },
                    marker: {
                        enabled: true,
                        hoveroptions: {
                            innerStrokeOpacity: 1,
                            outerFillColor: 'transparent',
                            outerStrokeWidth: 2
                        }
                    }
                },
                coloroverlay: {
                    show: false,
                    values: [],
                    overlayRange: 'below'
                },
                range: {
                    show: false,
                    values: [],
                    label: {
                        text: ['Description comes here'],
                        fontSize: userProp.fontL2Size.l5
                    }
                }
            }
        },
        yaxis: [
            {
                show: true,
                label: {
                    show: true,
                    text: 'Yaxis',
                    fontSize: userProp.fontL2Size.l4,
                    marginTop: 5,
                    marginRight: 5,
                    marginBottom: 5,
                    marginLeft: 5
                },
                grid: {
                    show: true,
                    color: '#ececec'
                },
                minorGrid: {
                    show: false,
                    color: '#ececec'
                },
                ticklabel: {
                    textOverflow: 'ellipsis', // none,ellipsis,hidden
                    fontSize: userProp.fontL2Size.l5
                },
                tickmark: {
                    color: '#ececec',
                    size: 3
                },
                subtickmark: {
                    color: '#ececec',
                    size: 5,
                    count: 0
                },
                axisline: {
                    show: true,
                    color: '#ececec',
                    strokeWidth: null,
                    style: 'solid'
                },
                baseline: {
                    show: false,
                    value: 0,
                    color: 'grey',
                    strokeWidth: null
                },
                threshold: {
                    line: {
                        show: false,
                        values: [],
                        label: {
                            text: ['Description comes here'],
                            fontSize: userProp.fontL2Size.l5
                        },
                        marker: {
                            enabled: true,
                            hoveroptions: {
                                innerStrokeOpacity: 1,
                                outerFillColor: 'transparent',
                                outerStrokeWidth: 2
                            }
                        }
                    },
                    coloroverlay: {
                        show: false,
                        values: [],
                        overlayRange: 'below'
                    },
                    range: {
                        show: false,
                        values: [],
                        label: {
                            text: ['Description comes here'],
                            fontSize: userProp.fontL2Size.l5
                        }
                    }
                }
            }
        ]
    };
    return axisOptions;
};

export var _axisOptions = options.axes;
