// @zohocharts/zc/src/axis/axis-events.js
//$Id$
//axis-events.js

import { validateNullVal, invokeFunction, pick, defined } from '../utils/zcutils';
import { TIME_DATATYPE } from '../core/datatype-constants';
import { generateDateVal } from '../formatter/formatter';
import { attachevents } from '../events/eventActions';
import { d3_event } from '../core/d3-constants';
import { ZC_AXIS_TICK_FILTER } from './tickFilter/tick-filter';
import { isBranchAxis } from './hierarchy-axis-utils';
import { getAxisObject } from './axisutils';
import { customGestureList } from '../events/custom-gesture';
import { Registry } from '../core/zc-registry';

export function bindEventHandlerForTicks(chartObj, tickG, ticklabels, tickImages, ticklabeldata, columnInfo, type, axisOrder) {
    var metaColumn = columnInfo,
        events_tick = ticklabeldata.events,
        eventDisabled = events_tick != null && validateNullVal(events_tick.enabled, true) == false;

    var axisObj = getAxisObject(chartObj, type, axisOrder),
        tickComp = axisObj && axisObj.ticks,
        tickObj = tickComp && tickComp.tickValueObj;

    var customHandlerFn = {
        mousemove: events_tick ? events_tick.mousemove : null,
        mouseout: events_tick ? events_tick.mouseout : null,
        click: events_tick ? events_tick.click : null,
        dblclick: events_tick ? events_tick.doubleclick : null,
        tap: events_tick ? events_tick.tap : null,
        tapHold: events_tick ? events_tick.taphold : null,
        tapHoldEnd: events_tick ? events_tick.tapholdend : null,
        dbltap: events_tick ? events_tick.doubletap : null
    };

    if (!eventDisabled) {
        var defaultHandlers = {
            click: defaultTickFilter,
            tap: defaultTickFilter
        };
        var callCustomHandler = (eventType, targetEvent, d, i, element) => {
            let callback = customHandlerFn[eventType] || (events_tick && events_tick[eventType]);
            if (callback) {
                invokeFunction(callback, targetEvent, getDatawithColumnInfo(columnInfo, d, i, axisOrder), chartObj, element);
            }
        };
        var mouseHandlers = {
            mousemove: function (e, d) {
                callCustomHandler('mousemove', d3_event(e), d, this.parentNode.__index__, this);
            },
            mouseout: function (e, d) {
                callCustomHandler('mouseout', d3_event(e), d, this.parentNode.__index__, this);
            },
            click: function(e, d) {
                callCustomHandler('click', d3_event(e), d, this.parentNode.__index__, this);
                if (!customHandlerFn.click || (customHandlerFn.click && e.allowDefault)) {
                    defaultHandlers.click.call(this, columnInfo, d, this.parentNode.__index__, axisOrder, false, type);
                }
            },
            dblclick: function (e, d) {
                callCustomHandler('dblclick', d3_event(e), d, this.parentNode.__index__, this);
            }
        };
        var touchHandlers = {
            tapHold: function (e, d) {
                callCustomHandler('tapHold', d3_event(e), d, this.parentNode.__index__);
            },
            tapHoldEnd: function (e, d) {
                callCustomHandler('tapHoldEnd', d3_event(e), d, this.parentNode.__index__);
            },
            tap: function(e, d) {

                callCustomHandler('tap', d3_event(e), d, this.parentNode.__index__);
                if (!customHandlerFn.tap || (customHandlerFn.tap && e.allowDefault)) {
                    defaultHandlers.tap.call(this, columnInfo, d, this.parentNode.__index__, axisOrder, false, type);
                }
            },
            dbltap: function (e, d) {
                callCustomHandler('dbltap', d3_event(e), d, this.parentNode.__index__);
            }
        };

        // cursor
        if (tickImages) {
            let cursor = eventDisabled ? 'auto' : events_tick ? events_tick.cursor || 'pointer' : 'pointer';
            tickImages.style('cursor', cursor);
        }

        attachevents(ticklabels, mouseHandlers, touchHandlers);
        if (tickImages) {
            attachevents(tickImages, mouseHandlers, touchHandlers);
        }

        if (tickObj && tickObj.customTickInfo && tickObj.customTickInfo.customGroup) {
            attachevents(tickG.selectAll('g.customTick'), mouseHandlers, touchHandlers);
        }

        if (customGestureList) {
            let gestureKeys = Object.keys(customGestureList),
                CustomGesture = Registry.getComponent('CustomGesture');
            gestureKeys.map(d => {
                CustomGesture.call(this, d, ticklabels, 'axis', callCustomHandler);
            });
            ticklabels = null;
        }
    }
    function defaultTickFilter(columnInfo, d, i, axisOrder, useDefault, type) {
        if (chartObj.dataObject.hierarchical && chartObj.dataObject.hierarchical.categoryTree) {
            var cType = chartObj.chartTypes[0],
                isBranchAxisType = isBranchAxis(type, axisOrder, chartObj);

            if ((cType.name === 'area' || cType.name === 'line') && !isBranchAxisType) {
                return;
            }

            var args = {
                event: d3_event(),
                columnInfo: getDatawithColumnInfo(columnInfo, d, i, axisOrder),
                axisType: type,
                element: this
            };

            ZC_AXIS_TICK_FILTER.call(chartObj, args);
            // invokeFunction(axis.TickLabelFilter, d3_event(), getDatawithColumnInfo(columnInfo, d, i, axisOrder), chartObj, true, type);
        }
    }

    function getDatawithColumnInfo(columnInfo, d, itemIndex, axisOrder) {
        //d.d => for new axis parser marks which is not implemented for polar axis
        columnInfo.data = columnInfo.data.map(function(dt) {
            let _data = defined(d) ? d.d || d : null;
            dt.value = dt.datatype === TIME_DATATYPE ? generateDateVal(d, null, dt.time).getTime() : _data;

            dt.itemIndex = itemIndex;
            dt.levelindex = pick(dt.levelindex, axisOrder);
            dt.axisOrder = axisOrder;

            return dt;
        });
        return columnInfo;
    }
    Registry.setComponent('AxisEventHandler', callCustomHandler);
}
