// @zohocharts/zc/src/axis/axis-init.js
//$Id$
//axis-init.js
import { axis as zcAxis, scale } from '../core/main-zc-init';
import { defined, getValByPriority, isArray, pick, utils, validateNullVal, arrayFind, invokeFunction } from '../utils/zcutils';
import { getAnimationConf } from '../utils/animationutils';
import { DataProcessor } from '../core/data-processor';
import {
    getAxisObject,
    getMargin,
    getTickPadding,
    isTickInnerOrient,
    merged_axis_column_intersect,
    getAxisData,
    isAxisEnabled,
    getAxesCount,
    getAxisLayout
} from './axisutils';
import { NUMERIC_DATATYPE, ORDINAL_DATATYPE, TIME_DATATYPE } from '../core/datatype-constants';
import { getDashArray } from '../utils/dashstyles';
import { appendEle, getTextContent, wrapTextContent } from '../utils/zcsvgutils';
import { mathAbs, mathMin, NONE, TRANSPARENT, UNDEFINED } from '../core/zc-constants';
import { dimension_based_intelligence } from '../utils/intelligence-utils';
import { handleOverflowTicks } from './ticklabel-overflow';
import { getFormattedValue, getFormattedValueByPercent } from '../formatter/formatter';
import { rendererBoundUtils } from '../plot/renderer-bound-utils';
import { $ZCG } from '../core/zcglobal';
import { hasFixedDatalabels } from '../label/checkFixedTopLabels';
import { d3_select } from '../core/d3-constants';
import { formatTickWrapByMode } from './tickWrap';
import { updateScaleByTransform } from '../behaviours/data-selection/chart/tools/transform-scale';
import { applyZoomTransform, createZoomScale } from '../behaviours/visual-effects/chart/tools/apply-zoom-transform-on-scale';
import { getOrdinalRange } from './scale';
import { fixBrushDirection } from '../behaviours/data-selection/chart/tools/selection-to-domain';
import { showAxisLabelForTree, getAxisDataIndex } from './hierarchy-axis-utils';
import { hasColorColumnInNonAxis } from '../utils/colorutils';
import { doTicksOverlaps } from './tick-helpers';
import { getAxisViewModel } from './axisutils';
import { createScaleComponetForParser } from '../parser/parser-add-scale';

export * from './dateinterval';
export * from './ticks';
export * from './tick-labels';

zcAxis.render = function(chartObj, chartWidth, chartHeight, update) {
    var systemconf = chartObj.systemConf,
        dataObject = chartObj.dataObject;
    var plotBorderDrawn = {
            left: false,
            right: false,
            top: false,
            bottom: false
        },
        wrap = chartObj.wrap,
        axes_animation_duration = get_animation_duration(),
        dataSplicer = chartObj.dataSplicer,
        metadata_axes = dataSplicer.getEncode(),
        metadata_columns = dataSplicer.getColumns(),
        yMajorTicksIndex = 0,
        xMajorTicksIndex = 0,
        // data processor helpers

        dataProcessorHelpers = DataProcessor.helpers,
        getParsedYAxisOrder = dataProcessorHelpers.getParsedYAxisOrder;

    chartObj.axes.animationDuration = axes_animation_duration;

    if (chartObj.legend.isSingleLayout) {
        // construct zscale
        chartObj.zscale = chartObj.dataObject.hasZaxis ? chartObj.scales[0].create() : null;
    }

    chartObj.datalabels.hasFixedPosition = chartObj.datalabels.moduleExists && hasFixedDatalabels(chartObj);
    chartObj.tickarea = {
        left: 0,
        right: 0,
        top: 0,
        bottom: 0
    };
    /*
     construct ticks
     */
    //copy of plotarea
    var plotarea_approx = {
        left: chartObj.plotarea.left,
        top: chartObj.plotarea.top,
        width: chartObj.plotarea.width,
        height: chartObj.plotarea.height
    };

    if (chartObj.dataObject.hierarchical) {
        var defaultRefDim_row = chartObj.axes.rotated ? 'width' : 'height',
            customDim_row =
                chartObj.systemConf.chart[defaultRefDim_row] == 'auto'
                    ? null
                    : chartObj.systemConf.chart[defaultRefDim_row] || chartObj.plotarea[defaultRefDim_row],
            defaultRefDim_col = chartObj.axes.rotated ? 'height' : 'width',
            customDim_col =
                chartObj.systemConf.chart[defaultRefDim_col] == 'auto'
                    ? null
                    : chartObj.systemConf.chart[defaultRefDim_col] || chartObj.plotarea[defaultRefDim_col];
        chartObj.dataObject.hierarchical.pivotLayout.leafSize(customDim_row, customDim_col, chartObj);
        // chartObj.dataObject.hierarchical.setLeafHeight();
        // if (!chartObj.axes.rotated) { // pivot12 - axisrotated throws error ZC2715
        scale.setRowScale(chartObj, 'row', true);
        // }
    }
    // if (chartObj.viewParser) {
    //
    // }
    zcAxis.constructTickObj(chartObj, chartWidth, chartHeight, 'left', plotarea_approx, update);
    zcAxis.constructTickObj(chartObj, chartWidth, chartHeight, 'bottom', plotarea_approx, update);

    clearCorrectionCache(chartObj);

    if (chartObj.dataObject.hierarchical) {
        var defaultRefDim_row = chartObj.axes.rotated ? 'width' : 'height',
            customDim_row =
                chartObj.systemConf.chart[defaultRefDim_row] == 'auto'
                    ? null
                    : chartObj.systemConf.chart[defaultRefDim_row] || chartObj.plotarea[defaultRefDim_row],
            defaultRefDim_col = chartObj.axes.rotated ? 'height' : 'width',
            customDim_col =
                chartObj.systemConf.chart[defaultRefDim_col] == 'auto'
                    ? null
                    : chartObj.systemConf.chart[defaultRefDim_col] || chartObj.plotarea[defaultRefDim_col];

        // chartObj.dataObject.hierarchical.findLabelWidth();
        chartObj.dataObject.hierarchical.pivotLayout.leafSize(customDim_row, customDim_col, chartObj);

        // chartObj.dataObject.hierarchical.setLeafHeight();
        scale.setRowScale(chartObj, 'row', true);
    }

    if (chartObj.legend.isSingleLayout) {
        chartObj.zscale = chartObj.dataObject.hasZaxis ? chartObj.scales[0].create() : null;
    }

    rendererBoundUtils.updatePlotareaByBandWidth(chartObj, plotarea_approx);

    // if(chartObj.customVisual){
    //     chartObj.viewflow.axes.vertical.map((axesObj, i) => {
    //         if(axesObj.scale.component){
    //             var axis_vertical = createAxis('left', i);
    //             axis_vertical = setAxisProp(axesObj.scale.component, 'left', i, update);
    //         }
    //     })
    // }
    //Left
    var scale1 = scale.constructor(chartObj, 'left');
    var type1 = chartObj.axes.rotated ? 'x' : 'y';

    // type = (orient == 'left' && chartObj.axes.rotated) || (orient == 'bottom' && !chartObj.axes.rotated) ? 'x' : 'y';
    var axesCount1 = getAxesCount(chartObj, type1, 'left');
    for (var i = 0; i < axesCount1; i++) {
        var axis_left = createAxis('left', i);
        let axisModel = getAxisViewModel(chartObj, type1, 'left', i),
            scaleComp = scale1[i].scale;
        if(axisModel && axisModel.scale){
            scaleComp = axisModel.scale.component;
        }
        if(chartObj.customVisual){
            if(axisModel && !axisModel.scale.component){
                scaleComp = axisModel.scale.component = createScaleComponetForParser(axisModel.scale.conf, chartObj);
            }
        }
        axis_left = setAxisProp(scaleComp, 'left', i, update);
    }

    //Bottom
    var scale2 = scale.constructor(chartObj, 'bottom');
    var type2 = chartObj.axes.rotated ? 'y' : 'x';
    var axesCount2 = getAxesCount(chartObj, type2, 'bottom');
    for (var i = 0; i < axesCount2; i++) {
        var axis_bottom = createAxis('bottom', i);
        let axisModel = getAxisViewModel(chartObj, type2, 'bottom', i),
            scaleComp = scale2[i].scale;
        if(axisModel && axisModel.scale){
            scaleComp = axisModel.scale.component;
        }
        if(chartObj.customVisual){
            if(axisModel && !axisModel.scale.component){
                scaleComp = axisModel.scale.component = createScaleComponetForParser(axisModel.scale.conf, chartObj);
            }
        }
        axis_bottom = setAxisProp(scaleComp, 'bottom', i, update);
    }
    

    if(chartObj.customVisual){
        for (var i = 0; i < axesCount1; i++) {
            var axisObj = getAxisObject(chartObj, type1, i);
            let axisModel = getAxisViewModel(chartObj, type1, 'bottom', i);
            // axisModel.component = 
            if (axisObj.component != null) {
                drawAxes(axisObj.component, i, type1, update);
            }
        }
        //draw axis at the end of creating axis prop
        for (var i = 0; i < axesCount2; i++) {
            var axisObj = getAxisObject(chartObj, type2, i);
            let axisModel = getAxisViewModel(chartObj, type2, 'bottom', i);
            if (axisObj.component != null) {
                drawAxes(axisObj.component, i, type2, update);
            }
        }
       
    }else{
        //Todo need to check the order flow
    //draw axis at the end of creating axis prop
    for (var i = 0; i < scale1.length; i++) {
        var axisObj = getAxisObject(chartObj, type1, i);
        if (axisObj.component != null) {
            drawAxes(axisObj.component, i, type1, update);
        }
    }
    //draw axis at the end of creating axis prop
    for (var i = 0; i < scale2.length; i++) {
        var axisObj = getAxisObject(chartObj, type2, i);
        if (axisObj.component != null) {
            drawAxes(axisObj.component, i, type2, update);
        }
    }
    }

    if (chartObj.dataset.getDataType('x') == ORDINAL_DATATYPE) {
        scale.barBandwidth(chartObj);
    }

    function get_animation_duration() {
        var axes_animation = systemconf.chart.axes.animation || {},
            enabled = axes_animation.enabled;

        if (!defined(enabled)) {
            var serdata =
                chartObj._isAxisCategory() || chartObj.isNonAxisMultiSeriesCategory ? chartObj.seriesdata : chartObj.seriesdata.slice(0, 1);
            enabled = serdata.some(function(d, i) {
                var type = chartObj.chartTypes[i].name;
                return getAnimationConf(systemconf.chart.plot.plotoptions[type], d.plotoptions, chartObj, type).enabled;
            });
        }

        return enabled ? pick(axes_animation.duration, 600) : 0;
    }
    //To bring axis line front
    function updateAxisLine(orient, axisdata, wrap, axisComp, isaxis, type, index) {
        // To bring axis line front, make plotarea border color as axis line color
        var plotBorderData = systemconf.chart.plot.border;

        var plotareaBorderEle = chartObj.staticPaneSVG.selectAll('g.plotareaBorder-property');
        var bkpInf = null;

        if (defined(plotBorderData) && plotBorderData.show === true) {
            chartObj.drawPlotBorder();
            wrap.selectAll('path.domain').style('stroke', TRANSPARENT);
        } else {
            var axislineData = axisdata.axisline;
            var axislineClr = axislineData.show == false || axislineData.color == null ? TRANSPARENT : axislineData.color;
            var axislineWidth = validateNullVal(axislineData.strokeWidth, 1);
            var axislineStyle = axislineData.style || 'solid';
            var isInline = isTickInnerOrient(axisdata, type, chartObj) && !axisComp.tickObj.tickValueObj.inlineToPlot;
            var tickLength = orient == 'left' || orient == 'right' ? axisComp.preCalcTickLen().width : axisComp.preCalcTickLen().height;
            var offset = isInline ? tickLength : 0;
            var axislineEle = appendEle(plotareaBorderEle, 'line', [1], '.', 'class', orient);

            if (isaxis == false || isAxisEnabled(axisdata, type, index, chartObj) == false || wrap.style('display') == NONE) {
                axislineEle.style('stroke', TRANSPARENT);
                return;
            }
            if (axislineData.zindex != 'back') {
                //=='front'
                var x1, x2, y1, y2;
                //0.5 correction for D3V4 translate
                if (orient == 'bottom') {
                    (x1 = -chartObj.tickarea.left),
                        (x2 = chartObj.plotarea.width + chartObj.tickarea.right),
                        (y1 = chartObj.plotarea.height + 0.5 + offset),
                        (y2 = chartObj.plotarea.height + 0.5 + offset);
                } else if (orient == 'left') {
                    (x1 = 0.5 - offset), (x2 = 0.5 - offset), (y1 = 0), (y2 = chartObj.plotarea.height + chartObj.tickarea.bottom);
                } else if (orient == 'right') {
                    (x1 = chartObj.plotarea.width + 0.5 + offset),
                        (x2 = chartObj.plotarea.width + 0.5 + offset),
                        (y1 = 0),
                        (y2 = chartObj.plotarea.height + chartObj.tickarea.bottom);
                } else if (orient == 'top') {
                    (x1 = -chartObj.tickarea.left),
                        (x2 = chartObj.plotarea.width + chartObj.tickarea.right),
                        (y1 = 0.5 - offset),
                        (y2 = 0.5 - offset);
                }

                axislineEle
                    .styles({
                        'stroke-width': axislineWidth,
                        stroke: axislineClr,
                        'shape-rendering': 'crispedges',
                        'stroke-dasharray': getDashArray(axislineStyle, axislineWidth)
                    })
                    .attrs({
                        x1: x1,
                        x2: x2,
                        y1: y1,
                        y2: y2
                    });

                bkpInf = {
                    size: axislineWidth,
                    clr: axislineClr
                };
                plotareaBorderEle.node()[`border-${orient}`] = bkpInf;

                wrap.selectAll('path.domain').style('stroke', TRANSPARENT);
            } else if (axislineData.zindex == 'back') {
                axislineEle.style('stroke', TRANSPARENT);
            }
        }
    }

    function getAxisOrient(data, type, index) {
        var orient = data.orient;
        var axesRotated = chartObj.axes.rotated,
            hierarchical = chartObj.dataObject.hierarchical,
            hasRows = chartObj.dataObject.categoryHierarchy && hierarchical.hasRows();
        if (orient) {
            return orient;
        } else if (type == 'x') {
            if (index == 0 || hierarchical) {
                orient = axesRotated ? 'left' : 'bottom';
            } else {
                orient = axesRotated ? 'right' : 'top';
            }
        } else {
            if (index == 0 || hasRows) {
                orient = axesRotated ? 'bottom' : 'left';
            } else {
                orient = axesRotated ? 'top' : 'right';
            }
        }
        return orient;
    }

    function createAxis(orient, index) {
        var axes = [];
        switch (orient) {
            case 'left':
                axes = chartObj.axes.rotated ? createAxisComp('x', index, orient) : createAxisComp('y', index, orient);
                break;
            case 'bottom':
                axes = chartObj.axes.rotated ? createAxisComp('y', index, orient) : createAxisComp('x', index, orient);
                break;
        }
        return axes;
    }

    function setAxisProp(scaleObj, orient, i, update) {
        var type = orient == 'left' ? (chartObj.axes.rotated ? 'x' : 'y') : chartObj.axes.rotated ? 'y' : 'x';
        var axisObj = getAxisObject(chartObj, type, i);

        return setAxisCompConfig(axisObj, type, scaleObj, orient, i, update);
    }

    function createAxisComp(type, index, orient) {
        var axes = getAxisObject(chartObj, type, index),
            axisDataOrder = getAxisDataIndex('y', index, chartObj);
        var axis_data = getAxisData(chartObj, type, index);

        // first axis disabled. on redraw it is enabled
        var axisComp = axes.component;
        if (axisComp) {
            return axisComp;
        }

        let show = isAxisEnabled(axis_data, type, index, chartObj);
        axes.component =
            type == 'x'
                ? show && dimension_based_intelligence(chartObj, 'xaxis', chartHeight)
                    ? zcAxis.get()
                    : null
                : show
                ? zcAxis.get()
                : null;

        return axisComp;
    }

    function drawAxes(axisObj, index, type, update) {
        var axes = [];
        axes = drawAxisComp(axisObj, type, index, update);
        return axes;
    }

    function setAxisCompConfig(axis, type, scaleObj, orient, index, update) {
        var axisObj = axis.component,
            hierarchyObj = chartObj.dataObject.hierarchical,
            hierarchical = hierarchyObj && hierarchyObj.categoryTree,
            hasMultiCluster = hierarchical ? hierarchyObj.pivotLayout.getRootNodeList('row').hasMultiCluster() : false,
            hasRows,
            isLastLeafScale,
            maxLeafAxisArea,
            axesObj = systemconf.chart.axes,
            axisdata = getAxisData(chartObj, type, index, false),
            // axisdata = type == 'x' ? axesObj.xaxis : axesObj.yaxis[getAxisDataIndex('y', index, chartObj)],
            orient = getAxisOrient(axisdata, type, index),
            dim = orient == 'bottom' || orient == 'top' ? chartHeight : chartWidth,
            isaxis = isAxisEnabled(axisdata, type, index, chartObj) === true && dimension_based_intelligence(chartObj, `${type}axis`, dim),
            ticklabelMargin = getMargin(axisdata.ticklabel),
            tickPadding = getTickPadding(orient, ticklabelMargin, isTickInnerOrient(axisdata, type, chartObj)),
            plotarea = chartObj.plotarea;

        if (hierarchical) {
            var leavesCount = hierarchyObj.pivotLayout.getRootNodeList('row').getLeavesCount();
            (hasRows = hierarchical ? hierarchyObj.hasRows() : null), (isLastLeafScale = hasRows && index == leavesCount - 1);
        }
        var ygroup = getAxisObject(chartObj, 'y', index);
        if (type === 'y' && (!isaxis || ygroup.disabled)) {
            yMajorTicksIndex++;
        }
        if (axisObj && !plotBorderDrawn[orient]) {
            var disabledYaxis = type === 'y' && ygroup.disabled;
            axisObj.plotBorderDrawn = !isaxis || disabledYaxis ? axisObj.plotBorderDrawn : true;
            plotBorderDrawn[orient] = axisObj.plotBorderDrawn;
        }

        // set it to the axis object
        axis.index = index;
        axis.reversed = axisdata.reversed;
        // clear zoom info before finding the base scale
        axis.baseScale = scale.getScale(chartObj, dataObject, plotarea, orient, index, chartObj.dataset, type, UNDEFINED, UNDEFINED, false);
        axis.scale.originalDomain = axis.baseScale.originalDomain;

        createZoomScale(axis, chartObj);
        applyZoomTransform(scaleObj, type, orient, axisdata.reversed, chartObj, axis);

        // #ZC2209
        scaleObj.ordinalRange = null;
        scaleObj.ordinalRange = getOrdinalRange(scaleObj);

        if (isaxis) {
            // if (1) {
            // //TODO AXIS PARSER Make all as separate utils. becz future teams may nee this too
            let axisObject = getAxisObject(chartObj, type, index),
                tickObj = axisObject.ticks,
                oldTickObj_preCalcAxisLen = tickObj.preCalcAxisLen,
                axis_prerender = axisObject.prerenderComp;
            axis_prerender.preRender(plotarea_approx);

            let { axisTicks, axisTitle } = axis_prerender;
            tickObj.tickValueObj = { ...axisTicks.tickInfo, ...axisTicks.precalculate, ...axisTitle.precalculate };
            tickObj.preCalcAxisLen = axisTitle.precalculate.preCalcAxisLen;
            tickObj.preCalcTickLen = axisTicks.precalculate.preCalcTickLen;

            var yaxisComp = chartObj.axes.y.groups,
                oldAxisLen = hasRows ? chartObj.axes[type].maxLeafAxisArea : oldTickObj_preCalcAxisLen,
                preCalcTickLen = axisTicks.tickInfo.longTickSize,
                adjustVal = tickObj.preCalcAxisLen - oldAxisLen,
                tickValueObj = tickObj.tickValueObj,
                canProcessAdjustment = hasRows ? isLastLeafScale : true,
                selector = type == 'x' ? `${type}-axis-${index}` : `${type}-axis-${index}`,
                axisWrap = appendEle(wrap.selectAll(`g.${type}-axis`), 'g', isaxis ? [axisdata] : [], '.', 'class', selector);
            axisWrap.attr('filter', null); //reset filter attribute on redraw

            // LOGx
            if (isArray(axisObject.logDomain)) {
                scaleObj.domain(axisObject.logDomain);
            }

            if (mathAbs(adjustVal) > 1 && canProcessAdjustment) {
                adjustPlotarea(orient, adjustVal);
                plotarea_approx = chartObj.plotarea;
            }

            if (axisObj != null) {
                var isMergedAxis = type == 'y' && !hierarchical ? metadata_axes.y[index].length > 1 : false,
                    hasTimegroupBy = tickObj && tickObj.isHierarchy() && chartObj.dataObject.hierarchical.groupByTime,
                    columnidx = chartObj.dataset.getColumnIdx(type, hasTimegroupBy ? 0 : index), //column index should be 0 for multi level x axis
                    columns = chartObj.dataSplicer.getColumns(),
                    datatype = columns[columnidx].datatype;
                datatype =
                    tickObj && tickObj.isHierarchialLevelType() && chartObj.dataObject.categoryHierarchy ? ORDINAL_DATATYPE : datatype;
                var formatConf = columns[columnidx][datatype];
                formatConf = isMergedAxis && type == 'y' ? merged_axis_column_intersect(index, chartObj) : formatConf;
                // formatConf.subfunction = axis_parser.subFormatType;
                var axisFormat = formatConf && (formatConf.axisformat || formatConf.format),
                    tickSize = validateNullVal(axisdata.tickmark.size, 6),
                    subTickSize = validateNullVal(axisdata.subtickmark.size, 6),
                    formattingFunction = dataObject.isPercentChart && type == 'y' ? getFormattedValueByPercent : getFormattedValue,
                    specifierIndex = hasTimegroupBy ? index : null,
                    isMultiLineTick =
                        (!(tickObj && tickObj.isHierarchy()) &&
                            formatConf &&
                            axisFormat &&
                            axisFormat.specifier &&
                            axisFormat.specifier.constructor == Array) ||
                        tickValueObj.mode == 'wrapOnSpace' ||
                        tickValueObj.mode == 'wrapAll';
                var isInline = tickValueObj.isInline && !tickValueObj.inlineToPlot,
                    w = orient === 'left' || orient === 'right' ? 'width' : 'height';
                chartObj.tickarea[orient] =
                    isInline && axisObj.plotBorderDrawn ? tickValueObj.axisgroup[w] : chartObj.tickarea[orient] || 0;
                axisObj
                    .tickOrient(tickValueObj.isInline ? 'inner' : 'outer')
                    .orient(orient)
                    .scale(scaleObj)
                    .tickPadding(tickPadding)
                    .tickSize(tickSize)
                    .subTickSize(subTickSize)
                    .tickSizeOuter(0)
                    .majorTicks(
                        index === (type === 'y' ? yMajorTicksIndex : xMajorTicksIndex) || hierarchical
                            ? getMajorTicks(orient, type, tickValueObj, index)
                            : null
                    )
                    .plotarea(chartObj.plotarea)
                    .systemconf(chartObj.systemConf)
                    .tickRotation(tickValueObj.rotation)
                    .preCalcTickLen(tickObj.preCalcTickLen)
                    .chartObj(chartObj)
                    .columnIndices(isMergedAxis && type == 'y' ? metadata_axes.y[index] : [columnidx])
                    .isMultiLineTick(isMultiLineTick)
                    .minorGrid(tickValueObj.minorTickValues);

                var format = formatConf != null ? formatConf.axisformat || formatConf.format : null;
                if (format && format.specifier == 'auto' && !scaleObj.bandwidth) {
                    axisObj.tickValues(tickObj.getTicksforAxisComp(tickValueObj)).tickFormat(function(d, i) {
                        args = [
                            null,
                            orient,
                            chartObj,
                            tickValueObj.rotation,
                            axisdata.ticklabel.textOverflow,
                            axisObj.scale(),
                            axisWrap,
                            axisObj,
                            'ticklabel',
                            axisdata.ticklabel,
                            index
                        ],
                        doOverflow = handleOverflowTicks.apply(null, args);
                        var x = scaleObj,
                            formattedValue = x.tickFormat()(d);
                        d3_select(this).text(formattedValue);
                        this.formattedVal = formattedValue;
                        doOverflow(this, i);
                    });
                } else {
                    axisObj.tickValues(tickObj.getTicksforAxisComp(tickValueObj));
                    var tickPos = tickObj.getActualTicklabel(tickValueObj);
                    var maxWidth = tickObj.tickValueObj.maxTickSize.horizontal,
                        maxHeight = tickObj.tickValueObj.maxTickSize.vertical,
                        fSize = utils.getVal(
                            getValByPriority([axisdata.ticklabel, chartObj.systemConf.chart, chartObj.systemConf.canvas, $ZCG], 'fontSize')
                        ),
                        needTickConfigWrap = tickObj.tickValueObj.mode == 'wrapOnSpace' || tickObj.tickValueObj.mode == 'wrapAll',
                        isMultiLine = axisObj.isMultiLineTick(),
                        isCustomLabel = defined(axisdata.ticklabel.showAs),
                        args = [
                            null,
                            orient,
                            chartObj,
                            tickValueObj.rotation,
                            axisdata.ticklabel.textOverflow,
                            axisObj.scale(),
                            axisWrap,
                            axisObj,
                            'ticklabel',
                            axisdata.ticklabel,
                            index
                        ],
                        doOverflow = handleOverflowTicks.apply(null, args);
                    var previousTick;
                    var tickFormat = function(d, i) {
                        var tick = tickPos(i);
                        var formattedValue = formattingFunction(formatConf, tick, datatype, chartObj, 'axis', null, null, specifierIndex);
                        d3_select(this).text(formattedValue);
                        this.formattedVal = formattedValue;
                        var isFirstTick = i == 0,
                            isLastTick = i == tickValueObj.ticks.length - 1;
                        let text_node;
                        

                        if (isMultiLine) {
                            if (needTickConfigWrap) {
                                formattedValue = formatTickWrapByMode(
                                    formattedValue,
                                    tickObj.tickValueObj.mode,
                                    maxWidth,
                                    maxHeight,
                                    fSize
                                );
                            }
                            wrapTextContent(d3_select(this), formattedValue, { orient: orient });
                        } else if (isCustomLabel) {
                            if (defined(tickObj.tickValueObj.customTicks[i])) {
                                this.innerHTML = '';
                                text_node = wrapTextContent(d3_select(this), tickObj.tickValueObj.customTicks[i], null, true);
                                text_node = text_node || this;
                                var textContent = getTextContent(d3_select(text_node));
                                text_node.formattedVal = isArray(textContent) ? textContent.join(' ') : textContent;
                            }
                        }

                        doOverflow(text_node  || this, i);
                        let minmaxlabels = axisdata.ticklabel.minmaxlabels,
                            continoustType =
                                (datatype == NUMERIC_DATATYPE || datatype == TIME_DATATYPE) && tickObj && !tickObj.isHierarchialLevelType();
                        if (continoustType && minmaxlabels && minmaxlabels.show && !minmaxlabels.excludeIntervals) {
                            if (i == 1) {
                                if (doTicksOverlaps(previousTick, this, tickValueObj.rotation, orient)) {
                                    this.style.display = 'none';
                                }
                            } else if (isLastTick) {
                                if (doTicksOverlaps(previousTick, this, tickValueObj.rotation, orient)) {
                                    previousTick.style.display = 'none';
                                }
                            }
                        }
                        previousTick = this;
                        // return formattedValue;
                    };
                    axisObj.tickFormat(tickFormat);
                }

                axisObj.tickObj = tickObj;
                axisObj.hideAxisLabel = !dimension_based_intelligence(chartObj, `${type}axislabel`, dim); //hide axislabel by intelligence
            }
        }

        return axisObj;
    }

    function drawAxisComp(axisComp, type, index, update) {
        var hierarchical = chartObj.dataObject.categoryHierarchy;
        var axesObj = systemconf.chart.axes,
            yaxisdataOrder = hierarchical ? getAxisDataIndex('y', index, chartObj) : index,
            axisdata = getAxisData(chartObj, type, index, false);
        // axisdata = type == 'x' ? axesObj.xaxis : axesObj.yaxis[yaxisdataOrder];
        var ticklabeldata = axisdata.ticklabel;
        var orient = getAxisOrient(axisdata, type, index);
        var dim = orient == 'bottom' || orient == 'top' ? chartHeight : chartWidth;
        var isaxis = isAxisEnabled(axisdata, type, index, chartObj) === true && dimension_based_intelligence(chartObj, `${type}axis`, dim);
        var selector = type == 'x' ? `${type}-axis-${index}` : `${type}-axis-${index}`;
        var axisWrap = appendEle(wrap.selectAll(`g.${type}-axis`), 'g', isaxis ? [axisdata] : [], '.', 'class', selector);
        var axisObj = getAxisObject(chartObj, type, index),
            axesCount = chartObj.axes[type].count;

        if (index > axesCount - 1) {
            appendEle(wrap.selectAll(`g.${type}-axis`), 'g', [], '.', 'class', selector);
            if (axisComp.plotBorderDrawn == true) {
                updateAxisLine(orient, axisdata, axisWrap, axisComp, false, type, index);
            }
            return;
        }
        var ygroup = getAxisObject(chartObj, type, index);
        //call adjust wrap here, without considering isAxis show or not (For multiY axis - toggle show option)
        axisWrap.style('display', type === 'y' && ygroup.disabled ? NONE : null);

        let strictAdjust = !chartObj.dataObject.isAxisCategory && index === chartObj.axes[type].count - 1;
        if (type === 'x' && isaxis) {
            adjustXAxiswrap(axisdata, index, axisComp, true, strictAdjust);
        } else if (isaxis) {
            adjustYAxeswrap(axisdata, index, axisComp, strictAdjust, true);
        }

        if (isaxis) {
            var tickObj = axisObj.ticks.tickValueObj;
            chartObj.plotarea = axisComp.plotarea();

            axisWrap
                .transition()
                .duration(axes_animation_duration)
                .call(axisComp, type, index);

            if (type == 'y' && axisComp.plotBorderDrawn == true) {
                updateAxisLine(orient, axisdata, axisWrap, axisComp, isaxis, type, index);
            } else if (type == 'x' && doUpdateAxisLine(chartObj, type, index)) {
                updateAxisLine(orient, axisdata, axisWrap, axisComp, isaxis, type, index);
            }
        } else {
            if (axisComp.plotBorderDrawn == true) {
                updateAxisLine(orient, axisdata, axisWrap, axisComp, isaxis, type, index);
            }
        }

        //Initializing chart brush before drawing x axes
        // let scrollSettings = getPropVal(systemconf.chart, 'scroll') || {};
        //
        // //Initializing chart brush before pre rrendering X tick obj
        // if (type === 'y' && scrollSettings.enabled === true) {
        //     chartObj.scroll.setPosition(chartObj.plotarea);
        // }
    }

    function doUpdateAxisLine(chartObj, type, index) {
        let _hierarchical = chartObj.dataObject.hierarchical;

        if (_hierarchical && !chartObj.dataObject.isAxisCategory && !hasColorColumnInNonAxis(chartObj)) {
            let nodeList = _hierarchical.pivotLayout.getRootNodeList(type === 'x' ? 'column' : 'row');
            return index === nodeList.getLeavesCount();
        } else {
            return index === 0;
        }
    }

    function adjustXAxiswrap(data, index, axis_comp, isDraw, strictMode) {
        //need to change name axis_comp and axisComp - code modularization
        var usedAxisSpace = {
            left: 0,
            top: 0,
            right: 0,
            bottom: 0
        };
        var axisCount = chartObj.axes.x.count;
        var hierarchical = chartObj.dataObject.categoryHierarchy,
            offset = 0;
        let hasMultiCluster, nodeListCount, leavesCount;
        if (hierarchical) {
            let pivotLayout = chartObj.dataObject.hierarchical.pivotLayout,
                nodeList = pivotLayout.getRootNodeList('column');
            nodeListCount = nodeList.getCount();
            hasMultiCluster = nodeList.hasMultiCluster();
            leavesCount = nodeList.getLeavesCount();
        }
        if (index == 0 || strictMode) {
            for (var i = 0; i < axisCount; i++) {
                let axisData = getAxisData(chartObj, 'x', i);

                if (isAxisEnabled(axisData, 'x', i, chartObj)) {
                    var axisObj = getAxisObject(chartObj, 'x', i);
                    var axisComp = axisObj.component;
                    var xorient = axis_comp.orient();
                    var xwrapboxWid = isDraw
                        ? getTickLabelWidth(data.label, data.ticklabel, axisComp.preCalcTickLen(), chartObj, 'x', i, xorient)
                        : axisComp.tickLabelWidth();
                    var xwrap = wrap.selectAll(`.x-axis-${i}`);
                    var isLeafScale = hierarchical && i < leavesCount - 1;
                    let axisView = getAxisViewModel(chartObj, 'x', xorient, i);
                    var transX =
                        xorient == 'right'
                            ? chartObj.plotarea.left + chartObj.plotarea.width + usedAxisSpace.right
                            : chartObj.plotarea.left - usedAxisSpace.left;
                    var transY =
                        chartObj.axes.rotated || xorient == 'top'
                            ? chartObj.plotarea.top - usedAxisSpace.top
                            : chartObj.plotarea.top + chartObj.plotarea.height + usedAxisSpace[xorient];
                    if(axisView && axisView.conf && axisView.conf.transform){
                        let transform = invokeFunction(axisView.conf.transform, [transX, transY], chartObj);
                        transX = transform[0];
                        transY = transform[1];
                    }else{
                        usedAxisSpace[xorient] += isLeafScale && i !== leavesCount - 1 ? 0 : xwrapboxWid;
                    }
                    
                    xwrap.attr('transform', `translate(${transX},${transY})`);
                    xwrap.attr('x', transX);
                    xwrap.attr('y', transY);

                    if (nodeListCount || hasMultiCluster) {
                        if (i == 0) {
                            offset = xwrapboxWid;
                        }

                        if (nodeListCount && i > leavesCount - 1) {
                            if (axisComp.majorTicks()) {
                                //axisrotated case need to handle
                                let attr = xorient == 'left' || xorient == 'right' ? 'x2' : 'y2',
                                    sign = xorient === 'top' || xorient === 'left' ? 1 : -1;
                                axisComp.majorTicks()[attr] += sign * offset;
                                offset += xwrapboxWid;
                            }
                        }
                    }
                }
            }
            var xwrap = wrap.selectAll('g.x-axis-0');

            // if (chartObj.axes.rotated) {
            //     xwrap
            //         .selectAll("line.majorTicks")
            //         .attr("x2", chartObj.plotarea.width)
            // }
            // updateAxisLine(xorient, data, xwrap, axis);
        }
    }

    function adjustYAxeswrap(data, index, axis_comp, strictMode, isDraw) {
        var yaxisData = systemconf.chart.axes.yaxis,
            yaxesLength = chartObj.axes.y.count;
        var yaxis0_wrap = wrap.selectAll('.y-axis-0');
        var yaxis0_orient = yaxisData[0].orient || (chartObj.axes.rotated ? 'bottom' : yaxisData[0].orient || 'left');
        var hierarchical = chartObj.dataObject.categoryHierarchy;
        if (strictMode || index == 0) {
            var usedAxisSpace = {
                left: 0,
                top: 0,
                right: 0,
                bottom: 0
            };
            var offset = 0;
            for (var j = 0; j < yaxesLength; j++) {
                let axisData = getAxisData(chartObj, 'y', j),
                    ydata = axisData,
                    show = isAxisEnabled(axisData, 'y', j, chartObj);
                if (show) {
                    var orient = ydata.orient;
                    orient =
                        orient ||
                        (chartObj.axes.rotated
                            ? j == 0 || hierarchical
                                ? 'bottom'
                                : 'top'
                            : ydata.orient != null
                            ? ydata.orient
                            : j == 0 || hierarchical
                            ? 'left'
                            : 'right'); //TODO use common getAxisOrient Func
                    var dim = orient == 'bottom' || orient == 'top' ? chartHeight : chartWidth;
                    var isaxis = show === true && dimension_based_intelligence(chartObj, 'y' + 'axis', dim);
                    if (isaxis) {
                        let axisView = getAxisViewModel(chartObj, 'y', orient, j);
                        let hasMultiCluster, nodeListCount, leavesCount;
                        if (hierarchical) {
                            let pivotLayout = chartObj.dataObject.hierarchical.pivotLayout,
                                nodeList = pivotLayout.getRootNodeList('row');
                            nodeListCount = nodeList.getCount();
                            hasMultiCluster = nodeList.hasMultiCluster();
                            leavesCount = nodeList.getLeavesCount();
                        }
                        var ygroup = getAxisObject(chartObj, 'y', j);
                        // var ygroup = hierarchical && (hierarchical.pivotLayout.rows.getCount() || hierarchical.pivotLayout.rows.hasMultiCluster()) ? j >  chartObj.axes.y.leafScales.length-1 ? chartObj.axes.y.rowGrids[j-chartObj.axes.y.leafScales.length] : chartObj.axes.y.leafScales[j] : chartObj.axes.y.groups[j];
                        var axisComp = ygroup.component;

                        var isLeafScale = hierarchical && j < leavesCount - 1;

                        var ywrap = wrap.selectAll(`.y-axis-${j}`);
                        var disabled = ywrap.style('display') == NONE;
                        var ywrapboxWid = disabled
                            ? axisComp.tickLabelWidth()
                            : getTickLabelWidth(ydata.label, ydata.ticklabel, axisComp.preCalcTickLen(), chartObj, 'y', j, orient);

                        var translateX =
                                orient == 'bottom' || orient == 'top'
                                    ? chartObj.plotarea.left
                                    : orient === 'left'
                                    ? chartObj.plotarea.left - usedAxisSpace[orient]
                                    : chartObj.plotarea.left + chartObj.plotarea.width + usedAxisSpace[orient],
                            translateY =
                                orient == 'left' || orient == 'right'
                                    ? chartObj.plotarea.top
                                    : orient === 'bottom'
                                    ? chartObj.plotarea.height + chartObj.plotarea.top + usedAxisSpace[orient]
                                    : chartObj.plotarea.top - usedAxisSpace[orient];
                        // if(isLeafScale && j > 0){
                        //     translateY += chartObj.axes.y.leafScales[j-1].scale.range()[1];
                        // }
                        if(axisView && axisView.conf && axisView.conf.transform){
                            let transform = invokeFunction(axisView.conf.transform, [translateX, translateY], chartObj);
                            translateX = transform[0];
                            translateY = transform[1];
                        }else{
                            usedAxisSpace[orient] += ygroup.disabled || (isLeafScale && j !== leavesCount - 1) ? 0 : ywrapboxWid;
                        }
                        

                        ywrap.attr('transform', `translate(${translateX},${translateY})`);

                        // handleOverflowTicks_init(ywrap, axisComp, chartObj, ydata, axisComp.orient());

                        //Bringing Axis line to front
                        if (axisComp && axisComp.plotBorderDrawn == true) {
                            updateAxisLine(orient, ydata, ywrap, axis_comp, 'y', j);
                        }

                        if (nodeListCount || hasMultiCluster) {
                            if (j == 0) {
                                offset = ywrapboxWid;
                            }

                            if (nodeListCount && j > leavesCount - 1) {
                                if (axisComp.majorTicks()) {
                                    //axisrotated case need to handle
                                    let attr = orient == 'left' || orient == 'right' ? 'x2' : 'y2',
                                        sign = orient === 'top' || orient === 'left' ? 1 : -1;
                                    axisComp.majorTicks()[attr] += sign * offset;
                                    offset += ywrapboxWid;
                                }
                            }
                        }
                    }
                }
            }

            !chartObj.axes.rotated
                ? yaxis0_wrap
                      .selectAll('line.majorTicks')
                      .attr('x2', yaxis0_orient == 'right' ? -chartObj.plotarea.width : chartObj.plotarea.width)
                : null;
        }
    }

    function getTickLabelWidth(labeldata, ticklabeldata, preCalcTickLen, chartObj, type, axisOrder, orient) {
        var axisLabelMargin = {
                top: 0,
                right: 0,
                bottom: 5,
                left: 0
            },
            curTickLabelWidth = 0,
            labelSize = 0,
            axisObj = getAxisObject(chartObj, type, axisOrder),
            axisComp = axisObj.component,
            tickComp = axisObj.ticks;

        if (type === 'clr') {
            labelSize = 0;
        } else {
            labelSize = axisObj.axisLabel.getBBox().height;
        }
        var hideAxisLabel =
            axisComp.hideAxisLabel ||
            labeldata.show === false ||
            !defined(labeldata.text) ||
            labeldata.text === '' ||
            (tickComp.isHierarchy() && !showAxisLabelForTree(type, axisOrder, chartObj));
        if (hideAxisLabel) {
            axisLabelMargin.top = axisLabelMargin.bottom = axisLabelMargin.left = axisLabelMargin.right = 0;
        } else {
            axisLabelMargin.top = pick(labeldata.marginTop, 5);
            axisLabelMargin.bottom = pick(labeldata.marginBottom, 5);
            axisLabelMargin.left = pick(labeldata.marginLeft, 5);
            axisLabelMargin.right = pick(labeldata.marginRight, 5);
        }

        var marginTickLabel_right = pick(ticklabeldata.marginRight, 5),
            marginTickLabel_left = pick(ticklabeldata.marginLeft, 5),
            marginTickLabel_bottom = pick(ticklabeldata.marginBottom, 5),
            marginTickLabel_top = pick(ticklabeldata.marginTop, 5);

        switch (orient) {
            case 'top':
                curTickLabelWidth = preCalcTickLen != null ? preCalcTickLen.height + marginTickLabel_top + axisLabelMargin.bottom : 0;
                curTickLabelWidth = curTickLabelWidth + labelSize + axisLabelMargin.top;
                break;
            case 'bottom':
                curTickLabelWidth = preCalcTickLen != null ? preCalcTickLen.height + marginTickLabel_bottom + axisLabelMargin.top : 0; //7 -> spacing between tickmark and ticklabels,axisLabelMargin.bottom -space bw ticklabel and axislabel
                curTickLabelWidth = curTickLabelWidth + labelSize + axisLabelMargin.bottom;
                break;
            case 'left':
                curTickLabelWidth = preCalcTickLen != null ? preCalcTickLen.width + marginTickLabel_left + axisLabelMargin.right : 0;
                curTickLabelWidth = curTickLabelWidth + labelSize + axisLabelMargin.left;
                break;
            case 'right':
                curTickLabelWidth = preCalcTickLen != null ? preCalcTickLen.width + marginTickLabel_right + axisLabelMargin.left : 0;
                curTickLabelWidth = curTickLabelWidth + labelSize + axisLabelMargin.right;
                break;
        }
        axisComp.tickLabelWidth(curTickLabelWidth);
        return curTickLabelWidth;
    }
    function getMajorTicks(orient, type, tickValueObj) {
        var x1, y1, x2, y2;
        x1 = y1 = x2 = y2 = 0;

        var isInline = tickValueObj.isInline && !tickValueObj.inlineToPlot,
            longTick = tickValueObj.longTickSize,
            inlineOffset = isInline ? (orient == 'left' || orient == 'right' ? longTick.width : 0) : 0; //adjust inline offfset only for left, right orient as gridline overlaps with  ticks

        if (orient == 'bottom') {
            y1 = inlineOffset;
            y2 = -chartObj.plotarea.height;
        } else if (orient == 'top') {
            y1 = -inlineOffset;
            y2 = chartObj.plotarea.height;
        } else if (orient == 'left') {
            x1 = -inlineOffset;
            x2 = chartObj.plotarea.width;
        } else if (orient == 'right') {
            x1 = inlineOffset;
            x2 = -chartObj.plotarea.width;
        }

        if (type === 'x' && chartObj.datalabels.hasFixedPosition) {
            var yaxiscolumnorder = chartObj.axes.y.groups
                    .map(function(d) {
                        return d.disabled;
                    })
                    .indexOf(false), // First enabled yaxis
                axisReversed = systemconf.chart.axes.yaxis[yaxiscolumnorder].reversed,
                correction = getGridLineCorrection(yaxiscolumnorder);

            if (orient == 'bottom') {
                axisReversed ? (y1 -= correction) : (y2 += correction);
            } else if (orient == 'top') {
                axisReversed ? (y2 -= correction) : (y1 += correction);
            } else if (orient == 'left') {
                axisReversed ? (x1 += correction) : (x2 -= correction);
            } else if (orient == 'right') {
                axisReversed ? (x2 += correction) : (x1 -= correction);
            }
        }

        return {
            x1: x1,
            y1: y1,
            x2: x2,
            y2: y2
        };
    }

    function getGridLineCorrection(index) {
        var yscale = scale.getScale(chartObj, chartObj.dataObject, chartObj.plotarea, chartObj.axes.y.orient, index, chartObj.dataset, 'y'),
            origRange = yscale.origRange,
            datatypeY = chartObj.dataset.getDataType('y', index),
            maxRange;

        if (datatypeY === ORDINAL_DATATYPE) {
            var domain = yscale.domain();
            maxRange = yscale(domain[domain.length - 1]) + yscale.bandwidth() / 2;
        } else {
            var domain = chartObj.dataObject.getMinMaxY(index, 0, chartObj.datalabels.hasFixedPosition);
            maxRange = yscale(domain[1]);
        }
        return mathMin(mathAbs(maxRange - origRange[0]), mathAbs(maxRange - origRange[1]));
    }

    function adjustPlotarea(orient, adjustVal) {
        var plotarea = chartObj.plotarea;
        if (orient === 'left') {
            plotarea.left += adjustVal;
            plotarea.width -= adjustVal;
        } else if (orient === 'right') {
            plotarea.width -= adjustVal;
        } else if (orient === 'bottom') {
            plotarea.height -= adjustVal;
            plotarea.bottom += adjustVal;
        } else {
            plotarea.height -= adjustVal;
            plotarea.top += adjustVal;
        }
    }

    function addorremoveAxis(axisWrap, axisComp, data, index, type, mode) {
        //Now only type is "Y" as multiY axis is available
        /*
         *For multiY axis charts alone,
         *mode - add, add the axis  mode - remove ,filtering or remove axis
         *Change the plotarea width, left with respect to the axis wrap width
         */
        //Reset plotborderdrawn and g.plotareaBorder-property
        plotBorderDrawn = {
            left: false,
            right: false,
            top: false,
            bottom: false
        };
        chartObj.staticPaneSVG
            .selectAll('g.plotareaBorder-property')
            .selectAll('line')
            .style('stroke', TRANSPARENT);
        var orient = axisComp.orient();
        var majorTicks = axisComp.majorTicks();
        var wrapWid = axisComp.tickLabelWidth();
        if (!(axisWrap.style('display') === 'none' && mode === 'remove')) {
            if (orient == 'left' || orient == 'right') {
                chartObj.plotarea.width = mode == 'add' ? chartObj.plotarea.width - wrapWid : chartObj.plotarea.width + wrapWid;
                chartObj.plotarea.left =
                    orient == 'left'
                        ? mode == 'add'
                            ? chartObj.plotarea.left + wrapWid
                            : chartObj.plotarea.left - wrapWid
                        : chartObj.plotarea.left;
            } else {
                chartObj.plotarea.height = mode == 'add' ? chartObj.plotarea.height - wrapWid : chartObj.plotarea.height + wrapWid;
                chartObj.plotarea.top =
                    orient == 'top'
                        ? mode == 'add'
                            ? chartObj.plotarea.top + wrapWid
                            : chartObj.plotarea.top - wrapWid
                        : chartObj.plotarea.top;
            }
        }

        var co = 0;
        var display = mode == 'add' ? '' : NONE;
        axisWrap.style('display', display);
        axisWrap.attr('filter', true); //To be used to check in updatescale func whether to call new tickObj or reuse it
        if (mode == 'add') {
            axisWrap.style('display', null);
        }

        for (var i = 0; i < chartObj.axes.y.groups.length; i++) {
            var ywrap = wrap.selectAll(`g.y-axis-${i}`);
            var curAxisComp = chartObj.axes.y.groups[i].component;
            var curAxisOrient = curAxisComp.orient();
            var curMajorTicks = curAxisComp.majorTicks();
            var curAxisData = systemconf.chart.axes.yaxis[i] || systemconf.chart.axes.yaxis[0];

            //Translate the yaxis wraps according to added/removed axiswrap width
            // adjustYAxeswrap(data, index, curAxisComp, true);

            if (ywrap.node()) {
                //If the axis to be removed have majorticks, then draw the major ticks for next upcoming axis
                if (ywrap.style('display') != NONE && co == 0) {
                    curMajorTicks = getMajorTicks(curAxisOrient, type, curAxisComp.tickObj.tickValueObj, i);
                    curAxisComp.majorTicks(curMajorTicks);
                    ywrap
                        .transition()
                        .duration(0)
                        .call(curAxisComp, type, i, true);
                    co++;
                } else if (curMajorTicks != null && co > 0) {
                    curAxisComp.majorTicks(null);
                    ywrap
                        .transition()
                        .duration(0)
                        .call(curAxisComp, type, i, true);
                }
                //set proper curAxisComp.plotBorderDrawn according to their visibility in order
                //for multi y axis on redraw need to check both visibility and display status
                if (ywrap.style('display') !== NONE && !plotBorderDrawn[curAxisOrient]) {
                    plotBorderDrawn[curAxisOrient] = true;
                    curAxisComp.plotBorderDrawn = true;
                } else {
                    curAxisComp.plotBorderDrawn = false;
                    //Set axisline color for the other components where plotBorderDrawn = false
                    if (ywrap.style('display') == '' || ywrap.style('display') == 'inline') {
                        ywrap.selectAll('path.domain').style('stroke', curAxisData.axisline.color);
                    } else {
                        ywrap.selectAll('path.domain').style('stroke', TRANSPARENT);
                    }
                }
            }
        }
        adjustYAxeswrap(data, index, axisComp, true);
    }
    chartObj.updateZscale = function() {
        if (chartObj.legend.isSingleLayout && chartObj.dataObject.hasZaxis) {
            let scaleComponent = arrayFind(chartObj.scales, _ => _.id === 'z-0'),
                domain = scaleComponent._parseDomain();
            chartObj.zscale.domain(domain);
        }
    };

    chartObj.updateScale = function(serData, type, zmInterval, axisorder, range, useExistingScale = false) {
        let order = (serData == null || isArray(serData) ? serData : serData.yaxiscolumnorder) || axisorder;

        var yaxiscolumnorder = type == 'y' ? getParsedYAxisOrder(order) : 0,
            axisorder = type == 'x' && defined(axisorder) ? axisorder : yaxiscolumnorder,
            axiswrap = wrap.selectAll(`g.${type}-axis-${axisorder}`),
            axes = getAxisObject(chartObj, type, axisorder),
            tickObj = axes.ticks,
            axisdata = getAxisData(chartObj, type, axisorder, false),
            ticklabeldata = axisdata.ticklabel,
            orient = axes.component ? axes.component.orient() : getAxisOrient(axisdata, type, yaxiscolumnorder),
            dim = orient == 'bottom' || orient == 'top' ? chartHeight : chartWidth,
            isaxis =
                isAxisEnabled(axisdata, type, type === 'x' ? 0 : yaxiscolumnorder, chartObj) === true &&
                dimension_based_intelligence(chartObj, `${type}axis`, dim),
            dataset = chartObj.dataset;

        var hierarchical = type == 'y' ? chartObj.dataObject.categoryHierarchy : chartObj.dataObject.hierarchical,
            columnidx = chartObj.dataset.getColumnIdx(type, yaxiscolumnorder),
            datatype = metadata_columns[columnidx].datatype,
            datatype =
                tickObj && tickObj.isHierarchialLevelType() && chartObj.dataObject.hierarchical.categoryTree ? ORDINAL_DATATYPE : datatype,
            formatConf = metadata_columns[columnidx][datatype],
            hasTimegroupBy = tickObj && tickObj.isHierarchy() && chartObj.dataObject.hierarchical.groupByTime,
            specifierIndex = hasTimegroupBy ? axisorder : null,
            getExtentVal = () => {
                if (chartObj.dataObject.pivot) {
                    return axes.scale.domain();
                }
                var cType = chartObj.chartTypes[0].name;
                var outerPadding = systemconf.chart.plot.plotoptions[cType].outerPadding;
                return dataset.getDataType(type, axisorder) === ORDINAL_DATATYPE
                    ? dataObject.getCategories(type, axisorder)
                    : type === 'x'
                    ? chartObj.dataObject.getMinMaxX(validateNullVal(outerPadding, 0.2))
                    : chartObj.dataObject.getMinMaxY(yaxiscolumnorder);
            },
            isNoData = false,
            plotarea = chartObj.plotarea,
            domain = zmInterval,
            newScale,
            behaviours = chartObj.behaviouralEvents.eventManager.behaviours,
            hasCustomBehaviours = Object.keys(behaviours).length > 0,
            hasPlotZoom = behaviours.PLOT_ZOOM && behaviours.PLOT_ZOOM.types && behaviours.PLOT_ZOOM.types[type],
            applySmoothEffect =
                hasPlotZoom || (behaviours.CHART_SCROLL && chartObj.scroll && chartObj.scroll.settings.smooth && type === 'x'),
            /**
             *
             * @param domain
             * @param applyZoom
             */
            getScale = (domain, applyZoom = true) => {
                newScale = scale.getScale(chartObj, dataObject, plotarea, orient, axisorder, dataset, type, domain, UNDEFINED, applyZoom);
                if (chartObj.dataObject.pivot) {
                    // for pivot chart scale should get already updated by scale.setRowScale, scale.setColumnScale
                    newScale = axes.scale;
                }
                newScale.ordinalRange = getOrdinalRange(newScale);
                return newScale;
            },
            // #ZC2225 $axisorder

            /**
             * updates the params of axis and scale to get a smooth effect
             * @param domain
             * @param range
             */
            updateZoomInfo = (domain, range) => {
                domain && axes.scale.domain((axes.domain = domain));
                range && axes.scale.range((axes.range = range));

                if (axes.scale.bandwidth) {
                    axes.scale.ordinalRange = null;
                }
                zmInterval = domain;
                axes.eventscale = axes.scale;
            };

        if (!zmInterval) {
            let t0 = getExtentVal() || [];
            isNoData = t0[0] === 'NoData' && t0[1] === 'NoData';
        }

        // when plot brushed to only one data. the bar inter padding is set to 0
        // when reseted the zoom  we need re-enable the padding
        // #ZC2172
        if (axes.scale.bandwidth && !applySmoothEffect && axes.domain && axes.domain.length === 1) {
            useExistingScale = false;

            // update the domain if available
            if (isArray(domain)) {
                axes.domain = domain;
            }
        }

        if (!useExistingScale) {
            // chartObj.cache.correction = {}; // bubblepie3 filter bar legend and undo
            // for ordinal smooth effect we don't change the domain for x axis
            if (type === 'x' && applySmoothEffect && axes.scale.bandwidth) {
                zmInterval = null;
            }

            // #ZC2845 CXZA_LOGAXIS_19
            if (axes.logDomain) {
                let logDomain = axes.logDomain;
                axes.logDomain = null;
                let extent = getExtentVal();
                if (extent[0] === 0 && extent[1] === 0) {
                    axes.logDomain = null;
                } else {
                    axes.logDomain = logDomain;
                }
            }

            newScale = getScale(zmInterval);
            axes.eventscale = axes.scale = newScale;

            // if zoom is there update, apply it
            if (applySmoothEffect) {
                // base scale should be the whole scale
                axes.baseScale = getScale(UNDEFINED, false);
                newScale.originalDomain = axes.baseScale.originalDomain;

                // get the transform
                let transform;
                if (hasPlotZoom) {
                    transform = behaviours.PLOT_ZOOM.instance.getTransform();
                } else {
                    transform = chartObj.scroll.transform;
                }

                if (transform) {
                    ({ domain, range } = updateScaleByTransform(axes, fixBrushDirection(type, chartObj.axes.rotated), transform));
                    // update the scale
                    updateZoomInfo(domain, range);
                }
            }
        }

        //Update scale
        if (useExistingScale) {
            updateZoomInfo(applySmoothEffect ? null : zmInterval, range);

            // set animation duration 0 for zoom
            chartObj.axes.animationDuration = chartObj.disableAnimation ? 0 : axes_animation_duration;
        }

        let removeAxis = type == 'y' && !axiswrap.empty() && isNoData,
            enableAxis =
                type == 'y' &&
                !axiswrap.empty() &&
                axisdata.disabled != true &&
                axiswrap.style('display') == NONE &&
                !!axiswrap.attr('filter');

        //Update axis
        if (removeAxis) {
            //Check again; no updation in ordinal y scale
            //No need to update the scale and axis
            if (isaxis) {
                addorremoveAxis(axiswrap, axes.component, axisdata, yaxiscolumnorder, 'y', 'remove');
            }
        } else if (enableAxis && !hasCustomBehaviours) {
            // currently while enabling axes, we do not update the tick labels.
            // for zoom we need to update it.
            if (isaxis) {
                addorremoveAxis(axiswrap, axes.component, axisdata, yaxiscolumnorder, 'y', 'add');
            }
        } else {
            if (dataObject.isBubbleEnabled && type == 'y' && !applySmoothEffect) {
                axes.eventscale = axes.scale = scale.getScale(
                    chartObj,
                    dataObject,
                    chartObj.plotarea,
                    orient,
                    yaxiscolumnorder,
                    chartObj.dataset,
                    'y'
                );
            }

            // when a axis is disabled and zoom the chart, the zoom effect does not apply on the disabled axis.
            // currently we are not updating the ticks for multi y axis while legend filter.
            // but if the chart is zoomed we should update the ticklabels once enabled the axis
            if (isaxis && hasCustomBehaviours && enableAxis) {
                addorremoveAxis(axiswrap, axes.component, axisdata, yaxiscolumnorder, 'y', 'add');
            }

            if (isaxis) {
                axes.component.scale(axes.scale);
                //set axisconfig
                var hasTimegroupBy = tickObj && tickObj.isHierarchy() && chartObj.dataObject.hierarchical.groupByTime,
                    columnidx = chartObj.dataset.getColumnIdx(type, hasTimegroupBy ? 0 : axisorder), //column index should be 0 for multi level x axis
                    datatype = metadata_columns[columnidx].datatype,
                    isMergedAxis = type == 'y' && !hierarchical ? metadata_axes.y[axisorder].length > 1 : false;
                datatype =
                    tickObj && tickObj.isHierarchialLevelType() && chartObj.dataObject.categoryHierarchy ? ORDINAL_DATATYPE : datatype;
                formatConf = metadata_columns[columnidx][datatype];
                formatConf = isMergedAxis && type == 'y' ? merged_axis_column_intersect(yaxiscolumnorder, chartObj) : formatConf;
                var formattingFunction = dataObject.isPercentChart && type == 'y' ? getFormattedValueByPercent : getFormattedValue;
                var formatConf_axis = formatConf != null ? formatConf.axisformat || formatConf.format : null;
                var tickObj = axes.ticks,
                    _oldpreCalcTickLen = tickObj.preCalcTickLen;
                var oldLongTickSize =
                    orient == 'bottom' || orient == 'top'
                        ? tickObj.tickValueObj.longTickSize.height
                        : tickObj.tickValueObj.longTickSize.width;

                let axis_prerender = axes.prerenderComp;
                axis_prerender.preRenderTickG(chartObj.plotarea);

                let { axisTicks, axisTitle } = axis_prerender;
                tickObj.tickValueObj = { ...axisTicks.tickInfo, ...axisTicks.precalculate, ...axisTitle.precalculate };
                tickObj.preCalcAxisLen = axisTitle.precalculate.preCalcAxisLen;

                tickObj.preCalcTickLen = _oldpreCalcTickLen; //TODO MUST CHANGE object from referenc; reuse tickobj referernce
                // formatConf.subfunction = axis_parser.subFormatType;
                var tickValueObj = tickObj.tickValueObj;
                // var tickValueObj = axes.prerenderComp.axisticks.tickInfo;
                // var tickValueObj = tickObj.getTickValues(
                //     chartObj.plotarea.width,
                //     chartObj.plotarea.height,
                //     zmInterval,
                //     false,
                //     null,
                //     axes.Ticks,
                //     chartObj.plotarea
                // );

                if (isArray(axes.logDomain) && !useExistingScale) {
                    axes.scale.domain(axes.logDomain);
                    axes.eventscale.domain(axes.logDomain);
                }

                tickObj.tickValueObj = tickValueObj;
                var curLongTickSize =
                    orient == 'bottom' || orient == 'top'
                        ? tickObj.tickValueObj.longTickSize.height
                        : tickObj.tickValueObj.longTickSize.width;
                var preCalcTickLen = tickObj.preCalcTickLen;

                var adjustVal = tickObj.retainAxisHeight ? 0 : 
                    oldLongTickSize > curLongTickSize
                        ? curLongTickSize - oldLongTickSize
                        : oldLongTickSize < curLongTickSize
                        ? curLongTickSize - oldLongTickSize
                        : 0;
                orient == 'bottom' || orient == 'top'
                    ? (preCalcTickLen.height = preCalcTickLen.height + adjustVal)
                    : (preCalcTickLen.width = preCalcTickLen.width + adjustVal);

                //FEATURESEVENTS_COMBO_7_1 -> reassigned adjusted tickObj.precalcTickLen to tickObj.tickValueObj.preCalcTickLen
                // tickObj.tickValueObj.preCalcTickLen.width = preCalcTickLen.width;
                // tickObj.tickValueObj.preCalcTickLen.height = preCalcTickLen.height;

                var fullUpation = !axiswrap.empty() && axiswrap.style('display') === NONE && type === 'y';
                if (fullUpation) {
                    axiswrap.style('display', null);
                }
                //set tick formats - multi line tick with format conf
                var tickValues = tickObj.getTicksforAxisComp(tickValueObj),
                    tickLabelWidth = getTickLabelWidth(
                        axisdata.label,
                        axisdata.ticklabel,
                        axes.component.preCalcTickLen(),
                        chartObj,
                        type,
                        axisorder,
                        orient
                    );
                axes.component
                    .tickValues(tickValues)
                    .tickRotation(tickValueObj.rotation)
                    .minorGrid(tickObj.tickValueObj.minorTickValues)
                    .preCalcTickLen(tickObj.preCalcTickLen);
                axes.component.tickLabelWidth(tickLabelWidth);
                var tickPos = tickObj.getActualTicklabel(tickValueObj);
                var maxWidth = tickObj.tickValueObj.maxTickSize.horizontal,
                    maxHeight = tickObj.tickValueObj.maxTickSize.vertical,
                    fSize = utils.getVal(
                        getValByPriority([axisdata.ticklabel, chartObj.systemConf.chart, chartObj.systemConf.canvas, $ZCG], 'fontSize')
                    ),
                    needTickConfigWrap = tickObj.tickValueObj.mode == 'wrapOnSpace' || tickObj.tickValueObj.mode == 'wrapAll',
                    isMultiLine = axes.component.isMultiLineTick(),
                    isCustomLabel = defined(axisdata.ticklabel.showAs),
                    args_tick = [
                        null,
                        orient,
                        chartObj,
                        tickValueObj.rotation,
                        axisdata.ticklabel.textOverflow,
                        axes.scale,
                        axiswrap, //	ZC2090
                        axes.component,
                        'ticklabel',
                        axisdata.ticklabel,
                        axisorder
                    ],
                    doOverflow = handleOverflowTicks.apply(null, args_tick);
                let previousTick;
                var tickFormat = function(d, i) {
                    var tick = tickPos(i);
                    var formattedValue = formattingFunction(formatConf, tick, datatype, chartObj, 'axis', null, null, specifierIndex);
                    d3_select(this).text(formattedValue);
                    var isFirstTick = i == 0,
                        isLastTick = i == tickValueObj.ticks.length - 1;
                    this.formattedVal = formattedValue;
                    if (isMultiLine) {
                        if (needTickConfigWrap) {
                            formattedValue = formatTickWrapByMode(formattedValue, tickObj.tickValueObj.mode, maxWidth, maxHeight, fSize);
                        }
                        wrapTextContent(d3_select(this), formattedValue, { orient: orient });
                    } else if (isCustomLabel) {
                        if (defined(tickObj.tickValueObj.customTicks[i])) {
                            this.innerHTML = '';
                            wrapTextContent(d3_select(this), tickObj.tickValueObj.customTicks[i], null, true);
                        }
                    }
                    doOverflow(this, i);
                    let minmaxlabels = axisdata.ticklabel.minmaxlabels,
                        continoustType =
                            (datatype == NUMERIC_DATATYPE || datatype == TIME_DATATYPE) && tickObj && !tickObj.isHierarchialLevelType();
                    if (continoustType && minmaxlabels && minmaxlabels.show && !minmaxlabels.excludeIntervals) {
                        if (i == 1) {
                            if (doTicksOverlaps(previousTick, this, tickValueObj.rotation, orient)) {
                                this.style.display = 'none';
                            }
                        } else if (isLastTick) {
                            if (doTicksOverlaps(previousTick, this, tickValueObj.rotation, orient)) {
                                previousTick.style.display = 'none';
                            }
                        }
                    }
                    previousTick = this;
                };

                if (formatConf_axis && formatConf_axis.specifier == 'auto' && !axes.scale.bandwidth) {
                    axes.component.tickFormat(function(d, i) {
                        d3_select(this).text(axes.scale.tickFormat()(d));
                        doOverflow(this, i);
                    });
                } else {
                    axes.component.tickFormat(tickFormat);
                }

                //call axis

                if (axes.component.isMultiLineTick() || defined(ticklabeldata.showAs)) {
                    axiswrap.call(axes.component, type, axisorder);
                } else {
                    axiswrap
                        .transition()
                        .duration(axes_animation_duration)
                        .call(axes.component, type, axisorder);
                }

                //Adjust wrap positions

                if (fullUpation) {
                    addorremoveAxis(axiswrap, axes.component, axisdata, yaxiscolumnorder, 'y', 'add');
                }

                if (mathAbs(adjustVal) > 1) {
                    //Bug IdZC795 //hack ; hav to change logic for redraw in zoom

                    if (!fullUpation) {
                        adjustPlotarea(orient, adjustVal);
                    }
                    var type_opp = type === 'x' ? 'y' : 'x',
                        oppositeAxis = type_opp === 'x' ? [chartObj.axes.x] : chartObj.axes.y.groups;

                    // #ZC2034
                    oppositeAxis.forEach(axisInfo => {
                        // #ZC2283 $!axisInfo.disabled
                        if (!axisInfo.disabled) {
                            if (applySmoothEffect) {
                                ({ domain, range } = axisInfo);
                                // update the range of the [type_opp]
                                chartObj.updateScale(serData, type_opp, domain, axisInfo.index, range, false);
                            } else {
                                chartObj.updateScale(serData, type_opp, null, axisInfo.index);
                            }
                        }
                    });
                    chartObj.drawPlotareaClippath();
                    chartObj.updatePlotareaProp();
                }
                if (type === 'y') {
                    axiswrap.styles({
                        display: null,
                        visibility: 'visible'
                    });
                    adjustYAxeswrap(axisdata, yaxiscolumnorder, axes.component, true);
                } else {
                    adjustXAxiswrap(axisdata, axisorder, axes.component);
                }

                var PW = chartObj.plotarea.width;
                !chartObj.axes.rotated && type == 'y'
                    ? wrap
                          .selectAll(`.y-axis-${yaxiscolumnorder}`)
                          .selectAll('line.majorTicks')
                          .attr('x2', orient != 'right' ? PW : -PW)
                    : null;
                if (axes.component.plotBorderDrawn == true) {
                    updateAxisLine(orient, axisdata, axiswrap, axes.component, type, axisorder);
                }
            }
        }
    };
};

function clearCorrectionCache(chartObj) {
    for (var key in chartObj.cache.correction.domainCorrection) {
        // if (!(key.indexOf('-y') > -1) && key !== 'measure') {
        //     chartObj.cache.correction.domainCorrection[key] = null;
        // }
    }
    for (var key in chartObj.cache.correction) {
        if (key !== 'domainCorrection') {
            chartObj.cache.correction[key] = null;
        }
    }
}
