// @zohocharts/zc/src/axis/axisutils.js
//$Id$
//axisutils.js

import { pick, intersectObject, mergeJSON } from '../utils/zcutils';
import { getLeafAxisOrder, isBranchAxis, getBranchDepth, isLeafAxis } from './hierarchy-axis-utils';
import { BOTTOM_ALIGN, LEFT_ALIGN } from '../core/align-constants';
import { hasColorColumnInNonAxis } from '../utils/colorutils';
import { ViewFlow } from '../parser/parser-flow';   //TODO: Move import to viewflow based rendering charts

export function merged_axis_column_intersect(axisorder, chartObj) {
    var metaData = chartObj.systemConf.metadata;
    var indices = metaData.axes.y[axisorder];
    var start = 1,
        end = indices.length - 1;

    var col1 = metaData.columns[indices[start - 1]];
    var output = col1[col1.datatype];

    for (; start <= end; start++) {
        var col2 = metaData.columns[indices[start]];
        var formatConf2 = col2[col2.datatype];

        output = intersectObject(output, formatConf2);
    }
    return output;
}

export function hasMergedAxis(helpers) {
    let { type, order, chart } = helpers,
        { dataObject, systemConf } = chart,
        dataSplicer = chart.dataSplicer,
        metadata_axes = dataSplicer.getEncode();
    return !dataObject.hierarchical && metadata_axes[type][order].length > 1;
}

export function getOppositeOrient(orient) {
    switch (orient) {
        case 'left':
            return 'right';
        case 'right':
            return 'left';
        case 'top':
            return 'bottom';
        case 'bottom':
            return 'top';
    }
}

/**
 *
 * @param {Object} obj
 * @param {number} defaultMargin
 * @returns {{top, right, bottom, left}}
 */

export function getMargin(obj, defaultMargin) {
    return {
        top: pick(obj.marginTop, defaultMargin, 5),
        right: pick(obj.marginRight, defaultMargin, 5),
        bottom: pick(obj.marginBottom, defaultMargin, 5),
        left: pick(obj.marginLeft, defaultMargin, 5)
    };
}

export function getTickPadding(orient, ticklabelMargin, invert) {
    if (!invert) {
        orient = getOppositeOrient(orient);
    }
    return ticklabelMargin[orient];
}

export function isRotatedChart(type, axisRotated) {
    return (type == 'x' && axisRotated) || (type == 'y' && !axisRotated);
}

export function getAxisType(orient, rotated) {
    if (((orient == 'bottom' || orient == 'top') && !rotated) || ((orient == 'left' || orient == 'right') && rotated)) {
        return 'x';
    } else {
        return 'y';
    }
}
export function getAxisObject(chartObj, type, index) {
    var obj;
    if (type == 'x') {
        let hierarchical = chartObj.dataObject.hierarchical;
        if (hierarchical) {
            let nodeList = hierarchical.pivotLayout.getRootNodeList('column'),
                nodeListCount = nodeList.getCount(),
                leavesCount = nodeList.getLeavesCount();

            if (nodeListCount && index > leavesCount - 1) {
                chartObj.axes.x.levels[index - leavesCount] = chartObj.axes.x.levels[index - leavesCount] || {};
                obj = chartObj.axes.x.levels[index - leavesCount];
            } else if (leavesCount > 1) {
                chartObj.axes.x.leaves[index] = chartObj.axes.x.leaves[index] || {};
                obj = chartObj.axes.x.leaves[index];
            } else {
                obj = chartObj.axes.x;
            }
        } else {
            obj = chartObj.axes.x;
        }
    } else if (type === 'clr') {
        obj = chartObj.axes.clr;
    } else {
        let hierarchical = chartObj.dataObject.hierarchical,
            nodeListCount,
            hasMultiCluster,
            leavesCount;
        if (hierarchical) {
            let nodeList = hierarchical.pivotLayout.getRootNodeList('row');
            nodeListCount = nodeList.getCount();
            hasMultiCluster = nodeList.hasMultiCluster();
            leavesCount = nodeList.getLeavesCount();
        }
        if (hasMultiCluster || nodeListCount) {
            // if(defined(chartObj.dataObject.hierarchical) && (chartObj.dataObject.hierarchical.rows.count || chartObj.dataObject.hierarchical.leafRows.hasMulti)){
            if (nodeListCount && index > leavesCount - 1) {
                chartObj.axes.y.branches[index - leavesCount] = chartObj.axes.y.branches[index - leavesCount] || {};
                obj = chartObj.axes.y.branches[index - leavesCount];
            } else {
                chartObj.axes.y.leaves[index] = chartObj.axes.y.leaves[index] || {};
                obj = chartObj.axes.y.leaves[index];
            }
        } else {
            chartObj.axes.y.groups[index] = chartObj.axes.y.groups[index] || {};
            obj = chartObj.axes.y.groups[index];
        }
    }
    return obj;
}

export function getAxisData(chartObj, type, index, needMerge) {
    var systemconf = chartObj.systemConf,
        axisConfig_new = chartObj.dataSplicer.getAxisConfig(type),
        axisconf = type == 'y' ? systemconf.chart.axes.yaxis : systemconf.chart.axes.xaxis,
        axisconf_levels = type == 'y' ? systemconf.chart.axes.yaxisLevels : systemconf.chart.axes.xaxisLevels,
        axisorder,
        processor = chartObj.dataObject,
        hierarchical = type == 'y' ? processor.categoryHierarchy : processor.hierarchical,
        categoryHierarchical = hierarchical && hierarchical.categoryTree,
        axisdata;

    if (hierarchical) {
        var isBranchAxisType = isBranchAxis(type, index, chartObj),
            leafAxisOrder = getLeafAxisOrder(index, null, type, chartObj);
        if (isBranchAxisType) {
            var branchDepth = getBranchDepth(type, index, chartObj);
            axisdata = axisconf_levels[branchDepth] || {};
            if (needMerge) {
                var basedata = type == 'y' ? axisconf[0] : axisconf;
                axisdata = axisconf_levels[branchDepth] = mergeJSON(axisconf_levels[branchDepth], basedata, false);
            }
        } else {
            axisdata = type == 'x' ? axisconf : axisconf[leafAxisOrder];
            if (needMerge && type == 'y' && leafAxisOrder > 0) {
                axisdata = mergeJSON(axisconf[leafAxisOrder], axisconf[0], false);
            }
        }
    } else {
        axisdata = chartObj.dataSplicer.getAxisConfig(type, index);
        if (needMerge && type == 'y' && index > 0) {
            axisdata = mergeJSON(axisconf[index], axisconf[0], false);
        }
    }
    return axisdata;
}

export function getAxesCount(chart, type, orient) {
    if (chart.viewFlow) {
        let layout = getAxisLayout(orient);
        if(chart.customVisual){
            return chart.viewFlow.axes[layout].length;
        }
        return chart.viewFlow.axes[layout].length || chart.axes[type].count;
    } else {
        return chart.axes[type].count;
    }
}

export function getAxisViewModel(chart, type, orient, idx) {
    let layout = getAxisLayout(orient);

    if (chart.viewFlow) {
        let axisView = chart.viewFlow.axes[layout];
        return axisView[idx];
    } else {
        return null;
    }
}

export function getAxisLayout(orient, type, rotated) {
    if (orient) {
        return orient == 'left' || orient == 'right' ? 'vertical' : 'horizontal';
    } else {
        if (rotated) {
            return type == 'x' ? 'vertical' : 'horizontal';
        } else {
            return type == 'y' ? 'vertical' : 'horizontal';
        }
    }
}

export function getScalesCount(chart, type, orient) {
    if (chart.viewFlow) {
        let layout = getAxisLayout(orient);
        return chart.viewFlow.scale[layout].length;
    } else {
        return chart.axes[type].count;
    }
}
export function isTickAutoRotate(chartObj, type, order, rotationConfig) {
    var isBranchAxisType;
    if (chartObj.dataObject.categoryHierarchy && type == 'x') {
        isBranchAxisType = isBranchAxis(type, order, chartObj);
    }
    return rotationConfig == 'auto' && !isBranchAxisType;
}

export function isTickInnerOrient(obj, type, chart) {
    if (obj.tickOrient == 'inner') {
        return type == 'x' && chart.ismultix ? false : true;
    }
}

export function checkTickInlineAlignWithInPlot(type, orient, systemConf) {
    var yaxisReversed = systemConf.chart.axes.yaxis[0].reversed;
    if (type == 'x') {
        if (
            (orient == 'bottom' && !yaxisReversed) ||
            (orient == 'left' && !yaxisReversed) ||
            (orient == 'top' && yaxisReversed) ||
            (orient == 'right' && yaxisReversed)
        ) {
            return true;
        }
    }
}

/**
 * Returns if the axis rotated
 * @returns {boolean}
 */
export function getAxisRotated(systemConf) {
    if(systemConf._viewParser){
        return systemConf.rotated;
    }else{
        var axes = systemConf.chart.axes;
        return axes && axes.rotated === true;
    }
    
}

export function getAxisAreaForLowerOrder(axisType, axisOrder, chartObj) {
    var hierachyObj = chartObj.dataObject.hierarchical,
        plotarea = chartObj.plotarea,
        attrObj,
        axesCount = chartObj.axes[axisType].count;
    var axisLen = 0;
    if (hierachyObj) {
        let tag = axisType == 'x' ? 'column' : 'row',
            nodeList = hierachyObj.pivotLayout.getRootNodeList(tag),
            leavesCount = nodeList.getLeavesCount(),
            branchCount = nodeList.getCount(),
            hasMultiCluster = nodeList.hasMultiCluster(),
            hasMultiLeaves = chartObj.dataObject.hierarchical && (hasMultiCluster || branchCount);
        for (var i = 0; i < axisOrder; i++) {
            // eslint-disable-next-line no-empty
            if (hasMultiLeaves && i < leavesCount - 1) {
            } else {
                let axisData = getAxisData(chartObj, axisType, i);
                if (isAxisEnabled(axisData, axisType, i, chartObj)) {
                    let axisObj = getAxisObject(chartObj, axisType, i);
                    axisLen +=
                        axisType === 'y' && i === leavesCount - 1 && hasMultiLeaves
                            ? chartObj.axes[axisType].maxLeafAxisArea
                            : axisObj.component.tickLabelWidth();
                }
            }
        }
    }
    return axisLen;
}

export function isAxisEnabled(config, axisType, i, chartObj) {
    let { show } = config;
    if (chartObj._isAxisCategory() && !chartObj.isAxisCategory) {
        let isPivot = chartObj.dataObject.hierarchical && chartObj.dataObject.hierarchical.pivot;
        if (isPivot) {
            if (isLeafAxis(axisType, i, chartObj) && (axisType === 'y' || !(axisType === 'x' && hasColorColumnInNonAxis(chartObj)))) {
                return false;
            }
        } else {
            return show;
        }
    }

    return show;
}

export function getAxisOrient(type, rotated, chartObj) {
    if (type == 'y') {
        return rotated ? BOTTOM_ALIGN : LEFT_ALIGN;
    } else {
        return rotated ? LEFT_ALIGN : BOTTOM_ALIGN;
    }
}

export function copyDefaultTickAttrs(source, target) {
    let customAnchor = target.attr('textAnchor');
    target.attrs({
        x: source.attr('x'),
        y: source.attr('y'),
        dy: source.attr('dy'),
        transform: source.attr('transform'),
        'text-anchor': source.attr('text-anchor'),
        class: source.attr('class'),
        style: source.attr('style')
    });
    if (customAnchor) {
        target.style('text-anchor', customAnchor);
    }
}
