// @zohocharts/zc/src/axis/axisComponent.js
//$Id$
//axisComponent.js

export * from './axis-init';
import { axis } from '../core/main-zc-init';
import { getAxisAreaForLowerOrder, getAxisData, getAxisObject } from './axisutils';
import { getClippathIdForAxis } from './axis-clippath';
import { bindEventHandlerForTicks } from './axis-events';
import { d3axis } from './d3-axis';
import { pick, defined, mergeJSON, isArray, invokeFunction } from '../utils/zcutils';
import { appendEle, addSVGtitleEle } from '../utils/zcsvgutils';
import { handleOverflowTicks } from './ticklabel-overflow';
import { scale } from '../core/main-zc-init';
import { getDashArray, getLineCap } from '../utils/dashstyles';
import { mathAbs, mathMin } from '../core/zc-constants';
import { d3_select, d3_axis } from '../core/d3-constants';
import { NONE } from '../core/zc-constants';
import { getAxisAreaDim, isBranchAxis, isLeafAxis } from './hierarchy-axis-utils';
import { getPositionForTicks } from './cartesian/ticks-position-finder';
import { renderDom } from '../mark/dom-render';
import { AxisRegistry } from './axis-registry';
import { ParserRegistry } from '../core/parser-registry';

axis.get = function() {
    var axisComp;
    var scaleObj,
        ticks,
        tickArguments = [],
        tickValues = null,
        tickFormat = null,
        tickSizeInner = 6,
        tickSizeOuter = 6,
        subTickSize = 6,
        tickPadding = 3,
        radius, axisType,
        axisRenderer,
        changeOfOrient,
        changeOfTickOrient;

    let newAxisParser;
    var majorTicks = null,
        plotarea,
        ticklabeldata,
        tickmarkdata,
        labeldata,
        griddata,
        minorGridData,
        subTickMarkData,
        tickLabelWidth = null,
        tickRotation = 0,
        systemconf,
        chartObj,
        preCalcTickLen,
        orient,
        tickOrient,
        columnIndices,
        isMultiLineTick,
        animationDuration,
        minorGrid,
        isPolar;

    function chart(selection, type, axisOrder, redraw) {
        selection.each(function(data) {
            var container = d3_select(this);

            labeldata = data.label;
            ticklabeldata = data.ticklabel;
            tickmarkdata = data.tickmark;
            griddata = data.grid;
            minorGridData = data.minorGrid || {};
            subTickMarkData = data.subtickmark || {};

            let dataSplicer = chartObj.dataSplicer,
            metadata_columns = dataSplicer.getColumns();

            var axisObj = getAxisObject(chartObj, type, axisOrder),
                tickComp = axisObj.ticks,
                tickObj = tickComp.tickValueObj,
                hierarchical = chartObj.dataObject.hierarchical,
                isCategoryhierarchical = chartObj.dataObject.categoryHierarchy,
                isBranchType = isCategoryhierarchical && isBranchAxis(type, axisOrder, chartObj),
                isGridShow = griddata.show,
                hasMinorGridPositions = isArray(minorGrid) && minorGrid.length > 0,
                isMinorGridShow = minorGridData.show && hasMinorGridPositions,
                isSubTickMarkShow = (subTickMarkData.count > 0 || defined(subTickMarkData.interval)) && hasMinorGridPositions,
                isHierarchialLevelType = tickComp.isHierarchialLevelType(),
                hasCustomTickLabel = defined(ticklabeldata.showAs);

            tickRotation = pick(tickRotation, 0);
            var rotation = tickRotation < 0 ? 360 - mathAbs(tickRotation) : tickRotation;
            rotation = rotation > 360 ? mathAbs(rotation) - 360 : rotation;

            if (changeOfOrient) {
                container
                    .selectAll('g.axisgroup')
                    .selectAll('*')
                    .remove(); //To retain axislabel we are removing only de generated axisgroup

                container.selectAll('g.axisgroup').attr('transform', null);

                container
                    .selectAll('g.axislabelgroup')
                    .attr('transform', null)
                    .attr('clip-path', null);
            }
            var g = appendEle(container, 'g', [data], '.', 'class', 'axisgroup');
            if (changeOfTickOrient) {
                container.selectAll(`g#ticklabelG_${chartObj.id}`).remove();
            }
            var tickLabelInlineG;
            if (tickObj && tickObj.inlineToPlot) {
                container
                    .selectAll('g.axisgroup')
                    .selectAll('*')
                    .remove();

                let gId = `ticklabelG_${chartObj.id}`;
                tickLabelInlineG = appendEle(container, 'g', [1], '#', 'id', gId);
                tickLabelInlineG.style('pointer-events', 'visible');
                var shadowEle = appendEle(chartObj.staticPaneSVG, 'use', [1]);
                shadowEle.attr('xlink:href', `#ticklabelG_${chartObj.id}`);
                shadowEle.attr('x', container.attr('x'));
                shadowEle.attr('y', container.attr('y'));
            }

            if (ticks) {
                axisComp.ticks(ticks[0], ticks[1]);
            }

            // add clip-path
            if (
                chartObj._isAxisCategory() &&
                type !== 'clr' &&
                !(!chartObj.dataObject.isAxisCategory && (hierarchical && hierarchical.pivot) && isLeafAxis(type, axisOrder, chartObj)) // Pivot non axis chart(Web chart) - no clip path for leaf axis
            ) {
                var id = getClippathIdForAxis(chartObj, tickObj, orient, type, axisOrder);
                g.attr('clip-path', `url(${chartObj.defsLocation}#${id})`);
                if (tickObj.inlineToPlot) {
                    tickLabelInlineG.attr('clip-path', `url(${chartObj.defsLocation}#${id})`);
                }
            }

            animationDuration = chartObj.axes.animationDuration;

            //call d3-axis
            var args = {
                tickValues: tickValues,
                tickFormat: tickFormat,
                tickSizeInner: tickSizeInner,
                tickSizeOuter: tickSizeOuter,
                subTickSize: subTickSize,
                gridSize: majorTicks,
                tickPadding: tickPadding,
                orient: orient,
                rotation: rotation,
                scale: scaleObj,
                chart: chartObj,
                axis: axisObj,
                minorGrid,
                syncTransition: chartObj.racing && chartObj.racing.syncTransition
            };
            tickComp.majorTicks = majorTicks;
            let transition_ = chartObj.racing && chartObj.racing.syncTransition || g.transition().duration(animationDuration);
            /*
            Use axis parser
             */
            let isLinear = isPolar ? false : true,
                preRenderC = axisObj.prerenderComp,
                positionH,
                axisTitleH,
                markInfo,
                markExtender,
                marks;
            if (isLinear) {
                let markParser = 'cartesian.default';
                if ( preRenderC.axisView ) {
                    markParser = preRenderC.axisView.conf.markParser || markParser ;
                    markExtender = preRenderC.axisView.conf.markExtender;
                }
                positionH = getPositionForTicks.call(axisObj.prerenderComp, true, scaleObj);
                axisTitleH = AxisRegistry.get('cartesian.axisTitle').call(axisObj.prerenderComp);
                markInfo = ParserRegistry.axis(markParser).call(preRenderC);
            }
            if (type === 'clr' || hasCustomTickLabel || isMultiLineTick) {
                args.renderGrid = true;
                args.renderTicklabel = tickObj.inlineToPlot ? false : true;
                args.isGridShow = isGridShow;

                if (isLinear) {
                    marks = markInfo(positionH, axisTitleH, args, transition_, g).mark;
                    if (markExtender) {
                        marks = invokeFunction(markExtender, marks, preRenderC);
                    }
                    renderDom(marks, g);
                } else {
                    if (isPolar) {
                        axisRenderer = axis.radial()
                            .axis(axisObj)
                            .chart(chartObj)
                            .scale(scaleObj)
                            .radius(radius)
                            .tickValues(tickValues)
                            .tickFormat(tickFormat)
                            .tickRotation(tickRotation)
                            .type(axisType);
                        newAxisParser = axisRenderer(g);
                    } else {
                        g.call(d3axis, args, data);
                    }
                }

                // g.call(d3axis, args, data);
                if (tickObj.inlineToPlot) {
                    args.renderGrid = false;
                    args.renderTicklabel = true;
                    marks = markInfo(positionH, axisTitleH, args, transition_, g).mark;
                    if (markExtender) {
                        marks = invokeFunction(markExtender, marks.mark, preRenderC);
                    }
                    renderDom(marks, tickLabelInlineG);
                    // tickLabelInlineG.call(d3axis, args, data);
                }
            } else {
                args.renderGrid = true;
                args.renderTicklabel = tickObj.inlineToPlot ? false : true;
                args.isGridShow = isGridShow;
                args.isMinorGridShow = isMinorGridShow;
                args.isSubTickMarkShow = isSubTickMarkShow;

                if (isLinear) {
                    marks = markInfo(positionH, axisTitleH, args, transition_, g).mark;
                    if (markExtender) {
                        marks = invokeFunction(markExtender, marks, preRenderC);
                    }
                    renderDom(marks, g);
                } else {
                    
                    if (isPolar) {
                        axisRenderer = axis.radial()
                            .axis(axisObj)
                            .chart(chartObj)
                            .scale(scaleObj)
                            .radius(radius)
                            .tickValues(tickValues)
                            .tickFormat(tickFormat)
                            .tickRotation(tickRotation)
                            .type(axisType);
                        newAxisParser = axisRenderer(g);
                    }
                    else {
                        g.transition(transition_).call(d3axis, args, data);
                    }
                }
                // g.transition(transition_).call(d3axis, args, data);
                if (tickObj.inlineToPlot) {
                    args.renderGrid = false;
                    args.renderTicklabel = true;
                    tickLabelInlineG
                        .transition()
                        .duration(150)
                        .call(d3axis, args, data);
                }
            }

            var tickG0 = g.selectAll('.tick');
            var ticks1 =
                tickValues == null
                    ? scaleObj.ticks
                        ? scaleObj.ticks.apply(scaleObj, [tickG0.nodes().length])
                        : scaleObj.domain()
                    : tickValues;
            let tickData = isLinear ? preRenderC.axisTicks.tickInfo.ticklabelObj : ticks1,
                tickDataAccessor = isLinear ? d => preRenderC.axisTicks.tickInfo.scale(d) : scaleObj;
            var tickG = g.selectAll('.tick').data(tickData, tickDataAccessor);
            var majorticksUpdate = tickG.select('line.majorTicks');
            var axisLabel = container.selectAll('text.axislabel');

            // Apply ticklabel properties

            var ticklabelG_ = tickObj.inlineToPlot ? tickLabelInlineG : g,
                ticklabels = ticklabelG_.selectAll('text.ticklabel').data(tickData, tickDataAccessor),
                tickImages = hasCustomTickLabel ? g.selectAll(' g.image').data(tickData, tickDataAccessor) : null;
            if (isHierarchialLevelType) {
                extendTickmarkLineLength(tickG, type, axisOrder, 'minorTicks');
                if (isBranchType && isGridShow) {
                    extendTickmarkLineLength(tickG, type, axisOrder, 'majorTicks');
                }
            }
            // var rotationInfo = getTickRotateInfo(orient, rotation, ticklabels, axisComp, tickmarkdata.size);

            var columnInfo = {
                data: columnIndices.map(function(d) {
                    return mergeJSON(
                        metadata_columns[d],
                        {
                            columnindex: d
                        },
                        true
                    );
                })
            };
            bindEventHandlerForTicks(chartObj, ticklabelG_, ticklabels, tickImages, ticklabeldata, columnInfo, type, axisOrder);

            //Add attribute level to the ticks
            if (type !== 'clr' && chartObj.dataObject.hierarchical && chartObj.dataObject.hierarchical.categoryTree) {
                setLevelAttributeForTick(tickG, ticklabels, axisLabel, type, axisOrder, columnInfo.data[0].datatype);
            }

            // Apply tickmarks properties
            if (!isHierarchialLevelType && tickmarkdata.align == 'between') {
                drawTickLinesBetweenTicklabels(scaleObj, tickG, 'minorTicks', type, axisOrder);
            }
            if (type !== 'clr') {
                axisLabel = container.selectAll('text.axislabel');
                let axisAreaDim = getAxisAreaDim(type, axisOrder, orient, chartObj, plotarea);
                // axisLabel.transformLabel(preCalcTickLen, axisAreaDim, orient);
                var axisLabelRotation = orient == 'left' ? 270 : orient == 'right' ? 90 : 0;
                handleOverflowTicks(axisLabel, orient, chartObj, axisLabelRotation, 'ellipsis', scaleObj, g, chart, 'axislabel');
                if (labeldata.tooltip === true) {
                    addSVGtitleEle(axisLabel, axisTitleH.text);
                }
            }

            if (majorTicks && isGridShow) {
                if ((scaleObj.bandwidth || scaleObj.discreteWidth) && tickmarkdata.align == 'between') {
                    drawTickLinesBetweenTicklabels(scaleObj, tickG, 'majorTicks', type, axisOrder);
                }
                if (type !== 'clr') {
                    //To bring axis line front, make grid lines overlapped with axis line transparent
                    var range = scaleObj.range();
                    var firstVal =
                        chart.tickValues() == null
                            ? null
                            : scaleObj(chart.tickValues()[0]) + (scaleObj.bandwidth ? scaleObj.bandwidth() / 2 : 0);
                    var plotBorderData = chartObj.systemConf.chart.plot.border,
                        type_opp = type === 'x' ? 'y' : 'x',
                        oppositeAxisData = getAxisData(chartObj, type_opp, 0, false);

                    var plotareaBorderColor = plotBorderData === true ? plotBorderData.color : oppositeAxisData.axisline.color;
                    if (
                        defined(firstVal) &&
                        griddata.color !== 'transparent' &&
                        plotareaBorderColor !== 'transparent' &&
                        range[0] === firstVal &&
                        !(type == 'y' && isCategoryhierarchical && hierarchical.hasRows())
                    ) {
                        d3_select(majorticksUpdate.nodes()[0]).style('stroke', 'transparent');
                    }
                }
            }

            //BaseLine
            var baseLineData = data.baseline;
            if (baseLineData != null) {
                var baseLine = appendEle(
                    g,
                    'line',
                    baseLineData.show === true && scaleObj.domain()[0] <= baseLineData.value && baseLineData.value < scaleObj.domain()[1]
                        ? [1]
                        : [],
                    '.',
                    'class',
                    'zeroBase'
                );
                baseLine
                    .attrs({
                        x1: orient == 'left' || orient == 'right' ? 0 : scaleObj(baseLineData.value),
                        y1: orient == 'left' || orient == 'right' ? scaleObj(baseLineData.value) : -plotarea.height,
                        x2: orient == 'left' || orient == 'right' ? plotarea.width : scaleObj(baseLineData.value),
                        y2: orient == 'left' || orient == 'right' ? scaleObj(baseLineData.value) : 0
                    })
                    .styles({
                        fill: NONE,
                        stroke: baseLineData.color,
                        'shape-rendering': 'crispedges',
                        'stroke-width': baseLineData.strokeWidth
                    });
            }
        });
        return chart;
    }

    chart.axis = axisComp;

    chart.scale = function(_) {
        if (!arguments.length) {
            return scaleObj;
        }
        scaleObj = _;
        if (axisComp) {
            axisComp.scale(scaleObj);
        } else {
            axisComp = d3_axis(orient, scaleObj);
            axisComp.scale(scaleObj);
            // d3_rebind(chart, axis, 'ticks', 'tickValues', 'tickSubdivide', 'tickSize', 'tickSizeOuter', 'tickPadding', 'tickFormat');
            // d3_rebind(chart, scale, 'domain', 'range', 'rangeBand', 'rangeBands', 'ticks');
        }

        return chart;
    };
    chart.systemconf = function(_) {
        if (!arguments.length) {
            return systemconf;
        }
        systemconf = _;
        return chart;
    };
    chart.orient = function(_) {
        if (!arguments.length) {
            return orient;
        }
        if (defined(orient)) {
            //Axis rotated in redraw, need to re-initiate the axis component
            if (orient != _) {
                changeOfOrient = true;
                chartObj.staticPaneSVG
                    .selectAll('g.plotareaBorder-property')
                    .selectAll('*')
                    .remove(); //#ZC1295
                axisComp = null;
            } else {
                changeOfOrient = false;
            }
        }
        orient = _;
        return chart;
    };

    chart.tickOrient = function(_) {
        if (!arguments.length) {
            return tickOrient;
        }
        if (defined(tickOrient)) {
            //Axis rotated in redraw, need to re-initiate the axis component
            if (tickOrient != _) {
                changeOfTickOrient = true;
            } else {
                changeOfTickOrient = false;
            }
        }
        tickOrient = _;
        return chart;
    };

    chart.tickArguments = function(_) {
        return arguments.length ? ((tickArguments = _ == null ? [] : Array.prototype.slice.call(_)), chart) : tickArguments.slice();
    };

    chart.tickValues = function(_) {
        return arguments.length
            ? ((tickValues = _ == null ? null : Array.prototype.slice.call(_)), chart)
            : tickValues && tickValues.slice();
    };

    chart.tickFormat = function(_) {
        return arguments.length ? ((tickFormat = _), chart) : tickFormat;
    };

    chart.tickSize = function(_) {
        return arguments.length ? ((tickSizeInner = tickSizeOuter = +_), chart) : tickSizeInner;
    };

    chart.subTickSize = function(_) {
        return arguments.length ? ((subTickSize = +_), chart) : subTickSize;
    };

    chart.tickSizeInner = function(_) {
        return arguments.length ? ((tickSizeInner = +_), chart) : tickSizeInner;
    };

    chart.tickSizeOuter = function(_) {
        return arguments.length ? ((tickSizeOuter = +_), chart) : tickSizeOuter;
    };

    chart.tickPadding = function(_) {
        return arguments.length ? ((tickPadding = +_), chart) : tickPadding;
    };
    chart.ticks = function() {
        if (!arguments.length) {
            return ticks;
        }
        ticks = arguments;
        return chart;
    };

    chart.majorTicks = function(_) {
        return arguments.length ? (majorTicks = _, chart) : majorTicks;
    };

    chart.minorGrid = function(_) {
        return arguments.length ? (minorGrid = _, chart) : minorGrid;
    };

    chart.plotarea = function(_) {
        return arguments.length ? (plotarea = _, chart) : plotarea;
    };

    chart.tickLabelWidth = function(_) {
        return arguments.length ? (tickLabelWidth = _, chart) : tickLabelWidth;
    };
    chart.tickRotation = function(_) {
        return arguments.length ? (tickRotation = _, chart) : tickRotation;
    };
    chart.preCalcTickLen = function(_) {
        return arguments.length ? (preCalcTickLen = _, chart) : preCalcTickLen;
    };
    chart.chartObj = function(_) {
        return arguments.length ? (chartObj = _, chart) : chartObj;
    };
    chart.columnIndices = function(_) {
        return arguments.length ? (columnIndices = _, chart) : columnIndices;
    };
    chart.isMultiLineTick = function(_) {
        return arguments.length ? (isMultiLineTick = _, chart) : isMultiLineTick;
    };

    chart.polar = function(_) {
        return arguments.length ? (isPolar = _, chart) : isPolar;
    };
    chart.radius = function (_) {
        return arguments.length ? (radius = _, chart) : radius;
    };
    chart.type = function (_) {
        return arguments.length ? (axisType = _, chart) : axisType;
    }
    chart.parser = function (_) {
        return arguments.length ? (newAxisParser = _, chart) : newAxisParser;
    }

    //To align the tick lines (grid lines or tick mark) in between the ticklabels
    function drawTickLinesBetweenTicklabels(scaleObj, tickG, type, axisType, axisOrder) {
        /*
        Remove old minor/major tickAlign elements
         */
        tickG.selectAll(`line#${type}Align`).remove();

        var x1,
            y1,
            x2,
            y2,
            ticks = tickG.nodes(),
            startTickG = d3_select(tickG.node()),
            lastTickG = d3_select(ticks[ticks.length - 1]),
            tickLineNode = tickG.selectAll(`line.${type}:not(#minorTicksAlign)`), // css selector is faster.
            lastTickLineNode = lastTickG.selectAll(`line.${type}`),
            newStartTickLineNode = appendEle(startTickG, 'line', [1], '#', 'id', `${type}Align`),
            axisObj = getAxisObject(chartObj, axisType, axisOrder),
            isHierarchialLevelType = axisObj.ticks.isHierarchialLevelType();

        var adjustDistance = getNewPositionForTickmark(scaleObj, isHierarchialLevelType);
        if (type == 'minorTicks') {
            (x2 = orient == 'bottom' || orient == 'top' ? 0 : -tickmarkdata.size),
                (y2 = orient == 'bottom' ? tickmarkdata.size : orient == 'top' ? -tickmarkdata.size : 0);
            applyStyleForTickmark(tickmarkdata, newStartTickLineNode);
        } else {
            var ticklabelWidth = getAxisAreaForLowerOrder(axisType, axisOrder, chartObj);
            var key = orient == 'left' || orient == 'right' ? 'x1' : 'y1',
                value = orient == 'left' || orient == 'top' ? ticklabelWidth : -ticklabelWidth;
            (x1 = key == 'x1' ? value : majorTicks.x1), (y1 = key == 'y1' ? value : majorTicks.y1);
            (x2 = majorTicks.x2), (y2 = majorTicks.y2);
            applyStyleForGrid(griddata, newStartTickLineNode);
        }

        if (scaleObj.bandwidth || scaleObj.discreteWidth) {
            tickLineNode.attr('transform', `translate(${adjustDistance.mid})`);
            lastTickLineNode.attr('transform', `translate(${adjustDistance.end})`);
            newStartTickLineNode.attr('transform', `translate(${adjustDistance.start})`).attrs({
                x1: x1,
                y1: y1,
                x2: x2,
                y2: y2
            });
        } else {
            var co = 0;
            tickLineNode.attr('transform', function(d, i) {
                var pos = getNewPositionForTickmark(scaleObj, isHierarchialLevelType, co++);
                return `translate(${pos.mid})`;
            });
            lastTickLineNode.attr('transform', `translate(${adjustDistance.end})`);
            newStartTickLineNode.attr('transform', `translate(${adjustDistance.start})`).attrs({
                x1: x1,
                y1: y1,
                x2: x2,
                y2: y2
            });
        }
    }

    function getNewPositionForTickmark(scaleObj, isHierarchialLevelType, index) {
        var startPos, midPos, endPos, value_1, value_2;
        if (scaleObj.bandwidth && !isHierarchialLevelType) {
            var rangeBand = scaleObj.bandwidth(),
                ordinalRanges = scaleObj.ordinalRange,
                outerPadding = scaleObj.outerPaddingPx + rangeBand / 2, //(ordinalRanges[0] - rangeExtent[0])+(rangeBand/2),
                diffBand = scaleObj.domain().length > 1 ? ordinalRanges[1] - ordinalRanges[0] - rangeBand : 0;
            (value_1 = rangeBand / 2 + diffBand / 2), (value_2 = mathMin(outerPadding, value_1));
        } else if (isHierarchialLevelType) {
            (value_1 = 0), (value_2 = 0);
        } else {
            let categories;
            if (chartObj.dataObject.hierarchical) {
                let nodelist = chartObj.dataObject.hierarchical.pivotLayout.getRootNodeList('column');
                categories = nodelist.getDomainForCluster(0);
            } else {
                categories = chart.tickValues();
            }
            // var categories = chartObj.dataObject.hierarchical ? chartObj.dataObject.hierarchical.leaves.key.list : chart.tickValues(),
            var continousRangeBand = scale.getContinuousRangeBand(scaleObj, categories, index),
                rangeBand = continousRangeBand.rangeBand;
            (value_1 = rangeBand), (value_2 = continousRangeBand.outerPaddingInPixel);
        }

        // midPos = (orient == "bottom" || orient == "top") ? (value_1 + ",0") : ("0,"+value_1),
        // startPos = (orient == "bottom" || orient == "top") ? ((-value_2) + ",0") : ("0,"+(-value_2)),
        // endPos = (orient == "bottom" || orient == "top") ? (value_2 + ",0") : ("0,"+value_2);

        //0.5 compensation for new change in D3V4 translate
        var adjustFactor = scaleObj.bandwidth ? 0.5 : 0;
        if (orient === 'bottom' || orient == 'top') {
            startPos = `${-(value_2 - adjustFactor)},0`;
            midPos = `${value_1 + adjustFactor},0`;
            endPos = `${value_2 + adjustFactor},0`;
        } else {
            startPos = `0,${-value_2 + adjustFactor}`;
            midPos = `0,${value_1 + adjustFactor}`;
            endPos = `0,${value_2 + adjustFactor}`;
        }

        return {
            start: startPos,
            mid: midPos,
            end: endPos
        };
    }

    // For hierarchy level type, tickmark line should get extend based on the last rendered axis ticklabel width

    function extendTickmarkLineLength(tickG, type, axisOrder, tickType) {
        var axisObj = getAxisObject(chartObj, type, axisOrder),
            axisComp1 = axisObj.component,
            tickConfig = tickType == 'minorTicks' ? tickmarkdata : griddata;
        var tickLineNode = tickG.selectAll(`line.${tickType}`);
        var ticklabelWidth = getAxisAreaForLowerOrder(type, axisOrder, chartObj);
        axisObj.ticks.tickValueObj.tickMarkSize = ticklabelWidth + axisComp1.tickSize();
        var index = axisOrder;

        var key = orient == 'left' || orient == 'right' ? 'x1' : 'y1',
            value = orient == 'left' || orient == 'top' ? ticklabelWidth : -ticklabelWidth;

        var nullKey = orient == 'left' ? 'y1' : 'x1';

        var pivotLayout = chartObj.dataObject.hierarchical.pivotLayout;
        var finalDepth = chartObj.axes[type].count - 1;
        if (index === finalDepth) {
            var y2 = orient == 'bottom' ? tickmarkdata.size : 0;
            var ticks = tickG.nodes();
            tickG.selectAll(`line#${tickType}Align`).remove();
            var endTickG = d3_select(ticks[ticks.length - 1]);
            var newEndTickmark = appendEle(endTickG, 'line', [1], '#', 'id', `${tickType}Align`);
            var scaleObj = axisObj.scale;
            var domainLen = scaleObj.domain().length;
            var toTranslate =
                orient == 'top' || orient == 'bottom'
                    ? `${scaleObj.discreteWidth(domainLen - 1)},0`
                    : `0,${scaleObj.discreteWidth(domainLen - 1)}`;
            var key2 = orient == 'left' || orient == 'right' ? 'x2' : 'y2',
                value2;
            if (tickType == 'minorTicks') {
                value2 = orient == 'left' || orient == 'top' ? -tickmarkdata.size : tickmarkdata.size;
            } else {
                value2 = orient == 'left' || orient == 'top' ? ticklabelWidth + plotarea.width : -(ticklabelWidth + plotarea.height);
            }

            newEndTickmark
                .attr('transform', `translate(${toTranslate})`)
                .attr(key, value)
                .attr(key2, value2);
            if (tickType == 'minorTicks') {
                applyStyleForTickmark(tickConfig, newEndTickmark);
            } else {
                applyStyleForGrid(tickConfig, newEndTickmark);
            }
        }

        tickLineNode.attr(nullKey, null);
        tickLineNode.attr(key, value);
    }

    function applyStyleForTickmark(data, ele) {
        ele.styles({
            fill: 'none',
            'shape-rendering': 'crispedges',
            'stroke-width': data.strokeWidth,
            stroke: data.color
        });
    }
    function applyStyleForGrid(data, ele) {
        var strokeWidth = pick(data.strokeWidth, 1),
            strokeDashArray = getDashArray(data.style, strokeWidth),
            strokeLineCap = getLineCap(data.lineCap, griddata.style);
        ele.styles({
            fill: 'none',
            'shape-rendering': 'crispedges',
            'stroke-width': strokeWidth,
            stroke: griddata.color,
            'stroke-dasharray': strokeDashArray,
            'stroke-linecap': strokeLineCap
        });
    }

    function setLevelAttributeForTick(tickG, ticklabels, axisLabel, type, axisOrder, datatype) {
        var pivotLayout = chartObj.dataObject.hierarchical.pivotLayout,
            tag = type == 'x' ? 'column' : 'row',
            nodeList = pivotLayout.getRootNodeList(tag),
            isBranchType = isBranchAxis(type, axisOrder, chartObj),
            nodes,
            nodeIdx;
        if (isBranchType) {
            var branchIdx = axisOrder - nodeList.getLeavesCount(),
                nodes = nodeList.getBranchNodes(branchIdx);
        } else {
            var hasSplitLeafAxis =
                (tag == 'column' && nodeList.splitColumnLeaves && chartObj.axes[type].leaves.length) ||
                (tag == 'row' && datatype == 'numeric');
            nodes = hasSplitLeafAxis ? nodeList.getBranchNodes(0) : nodeList.getClusterNodes(0);
            nodeIdx = hasSplitLeafAxis ? chartObj.axes[type].leaves[axisOrder].parentBranchIdx : null;
        }
        if (nodes) {
            tickG
                .attr('levelbycolumn', function(d, i) {
                    var idx = pick(nodeIdx, i);
                    return nodes[idx] ? nodes[idx].levelByColumn : null;
                })
                .attr('levelbyrow', function(d, i) {
                    var idx = pick(nodeIdx, i);
                    return nodes[idx] ? nodes[idx].levelByRow : null;
                });

            if (datatype == 'numeric') {
                axisLabel.attr('levelbycolumn', nodes[nodeIdx || 0].levelByColumn);
                axisLabel.attr('levelbyrow', nodes[nodeIdx || 0].levelByRow);
            }
            var cache = chartObj.cache.hierarchical && chartObj.cache.hierarchical[tag],
                key = tag == 'row' ? 'levelByRow' : 'levelByColumn';
            if (cache && cache.active) {
                var query = cache.query;
                ticklabels.style('opacity', (d, i) => {
                    var idx = pick(nodeIdx, i),
                        activeTick = nodes[idx][key].startsWith(query);
                    if (activeTick) {
                        return 1;
                    } else {
                        return 0.1;
                    }
                });
            }
        }
    }

    return chart;
};

export var _axisRenderer = axis.get;
