// @zohocharts/zc/src/axis/d3-axis.js
//$Id$
//d3-axis.js
import { validateNullVal, pick, defined, isWithinRange, getValByPriority, utils, invokeFunction } from '../utils/zcutils';
import { getDashArray, getLineCap } from '../utils/dashstyles';
import { addSVGtitleEle, createElement } from '../utils/zcsvgutils';
import { $ZCG } from '../core/zcglobal';
import { getTickRotateInfo } from './tickRotate-helpers';
import { centeringTicklabel } from './centeringTicks';
import { updateMinmaxlabelPosition } from './minmaxLabelPosition';
import { d3_select } from '../core/d3-constants';
import { getSvgTransform } from '../utils/svg-transforms';

var identity$1 = function(x) {
    return x;
};
function translateX(x) {
    return `translate(${x + 0.5},0)`;
}

function translateY(y) {
    return `translate(0,${y + 0.5})`;
}

function number$1(scale1) {
    return function(d) {
        return +scale1(d);
    };
}

function center(scale1) {
    var offset = Math.max(0, scale1.bandwidth() - 1) / 2; // Adjust for 0.5px offset.
    if (scale1.round()) {
        offset = Math.round(offset);
    }
    return function(d) {
        return +scale1(d) + offset;
    };
}

function center1(scale1) {
    return function(d, i) {
        var offset = Math.max(0, scale1.bandwidth(d) - 1) / 2; // Adjust for 0.5px offset.
        // if (scale1.round()) {
        offset = Math.round(offset);
        // }
        return +scale1(d) + offset;
    };
}

function entering() {
    return !this.__axis;
}

//! Migrate to parser code
export function d3axis(context, args, data) {
    var tickArguments = [],
        chart = args.chart,
        axis = args.axis,
        tickObj = axis.ticks.tickValueObj,
        tickValues = args.tickValues,
        tickFormat = args.tickFormat,
        tickSizeInner = args.tickSizeInner,
        tickSizeOuter = args.tickSizeOuter,
        subTickSize = args.isSubTickMarkShow ? args.subTickSize : 0,
        gridSize = args.gridSize,
        tickPadding = args.tickPadding,
        rotation = args.rotation,
        orient = args.orient,
        scale1 = args.scale,
        isInline = tickObj && tickObj.isInline,
        isInwardPlot = tickObj && tickObj.inlineToPlot,
        k = orient === 'top' || orient === 'left' ? -1 : 1,
        l = isInwardPlot ? -k : k,
        m = isInline ? -k : k,
        x = orient === 'left' || orient === 'right' ? 'x' : 'y',
        w = orient === 'left' || orient === 'right' ? 'width' : 'height',
        y = x == 'x' ? 'y' : 'x',
        transform = orient === 'top' || orient === 'bottom' ? translateX : translateY,
        h = chart.dataObject.isPolarAxisCategory ? 'H' : 'h',
        labeldata = data.label,
        ticklabeldata = data.ticklabel,
        events_tick = ticklabeldata.events,
        tickmarkdata = data.tickmark,
        subtickmark = data.subtickmark,
        axisline = data.axisline,
        griddata = data.grid,
        minorGridData = data.minorGrid || {},
        tickTransform = data.tickTransform,
        renderGrid = args.renderGrid,
        renderTicklabel = args.renderTicklabel,
        showGrid = args.isGridShow,
        showMinorGrid = args.isMinorGridShow,
        isSubTickMarkShow = args.isSubTickMarkShow,
        eventDisabled = events_tick != null && validateNullVal(events_tick.enabled, true) == false,
        cursor = eventDisabled ? 'auto' : events_tick ? events_tick.cursor || 'pointer' : 'pointer',
        plotBorderData = chart.systemConf.chart.plot.border,
        minorGridValues = args.minorGrid,
        isLevelType = axis.ticks.isHierarchialLevelType();

    var epsilon = 1e-6;

    var values = tickValues == null ? (scale1.ticks ? scale1.ticks.apply(scale1, tickArguments) : scale1.domain()) : tickValues,
        format = tickFormat == null ? (scale1.tickFormat ? scale1.tickFormat.apply(scale1, tickArguments) : identity$1) : tickFormat,
        spacing = isInline && !isInwardPlot ? tickObj.longTickSize[w] : Math.max(tickSizeInner, subTickSize, 0) + tickPadding,
        tickmarkStart = isInline && !isInwardPlot ? tickObj.axisgroup[w] : 0,
        range = scale1.origRange ? scale1.origRange : scale1.range(),
        range0 = +range[0] + 0.5,
        range1 = +range[range.length - 1] + 0.5,
        position = (scale1.bandwidth ? (scale1.discreteWidth ? (isLevelType ? number$1 : center1) : center) : number$1)(scale1.copy()),
        selection = context.selection ? context.selection() : context,
        path = selection.selectAll('.domain').data([null]),
        tick = selection
            .selectAll('.tick')
            .data(values, scale1)
            .order(),
        tickExit = tick.exit(),
        tickEnter = tick
            .enter()
            .append('g')
            .attr('class', 'tick'),
        minorTicks = tick.select('line.minorTicks'),
        majorTicks = gridSize && griddata.show ? tick.select('line.majorTicks') : tick.select('line.majorTicks').remove(),
        ticklabel = tick.select('text').attr('class', 'ticklabel');

    // minor grid
    showMinorGrid = showMinorGrid && isArray(minorGridValues);
    isSubTickMarkShow = isSubTickMarkShow && isArray(minorGridValues);

    let subGridGroup = createElement('g', '.minorGrid', showMinorGrid ? [1] : [], selection);
    let minorGrids = createElement('line', '.minor-grid', showMinorGrid ? minorGridValues : [], subGridGroup);

    // minor grid
    let subTicksGroup = createElement('g', '.subTicks', isSubTickMarkShow ? [1] : [], selection);
    let subTicks = createElement('line', '.sub-ticks', isSubTickMarkShow ? minorGridValues : [], subTicksGroup);

    let transition_ = chart.racing && chart.racing.syncTransition || selection.transition(context);

    path = path
        .merge(path.enter().insert('path', '.tick'))
        .attr('class', 'domain')
        .styles({
            stroke: axisline.show === false ? 'transparent' : axisline.color || '#000',
            'stroke-width': pick(axisline.strokeWidth, 1),
            'stroke-dasharray': getDashArray(axisline.style, pick(axisline.strokeWidth, 1)),
            'stroke-linecap': getLineCap(axisline.lineCap, axisline.style)
        });

    tick = tick.merge(tickEnter);
    tick.property('__index__', (_, i) => i)     //* For item index storage

    const applyMinorTickStyles = (selection, config, size, applyTransform) => {
        if (applyTransform) {
            selection.attr('transform', d => transform(position(d)));
        }

        selection
            .attr(`${x}1`, k * tickmarkStart)
            .attr(`${x}2`, k * tickmarkStart + m * size)
            .styles({
                fill: 'none',
                'shape-rendering': 'crispedges',
                'stroke-width': config.strokeWidth,
                stroke: config.color
            });
    };

    // #ZC1998
    minorTicks = createElement('line', '.minorTicks', [1], tick);
    applyMinorTickStyles(minorTicks, tickmarkdata, tickSizeInner);

    //
    // .attr(`${x}1`, k * tickmarkStart)
    // .attr(`${x}2`, k * tickmarkStart + m * tickSizeInner)
    // .styles({
    //     fill: 'none',
    //     'shape-rendering': 'crispedges',
    //     'stroke-width': tickmarkdata.strokeWidth,
    //     stroke: tickmarkdata.color
    // });

    if (subtickmark) {
        applyMinorTickStyles(subTicks, subtickmark, subTickSize, true);
    }

    // subTicks
    //     .attr('transform', d => transform(position(d)))
    //     .attr(`${x}1`, k * tickmarkStart)
    //     .attr(`${x}2`, k * tickmarkStart + m * minorTickSize)
    //     .styles({
    //         fill: 'none',
    //         'shape-rendering': 'crispedges',
    //         'stroke-width': minorTickmarkdata.strokeWidth,
    //         stroke: minorTickmarkdata.color
    //     });

    //
    const applyGridStyles = (selection, config) => {
        var strokeWidth = pick(config.strokeWidth, 1),
            strokeDashArray = getDashArray(config.style, strokeWidth),
            strokeLineCap = getLineCap(config.lineCap, config.style);

        // #ZC1998
        selection
            .attr('stroke', '#000')
            .attrs({
                x1: gridSize.x1,
                y1: gridSize.y1,
                x2: gridSize.x2,
                y2: gridSize.y2
            })
            .styles({
                fill: 'none',
                'shape-rendering': 'crispedges',
                'stroke-width': strokeWidth,
                stroke: config.color,
                'stroke-dasharray': strokeDashArray,
                'stroke-linecap': strokeLineCap
            });
    };

    if (renderGrid && gridSize && showGrid) {
        // apply major grid styles
        majorTicks = createElement('line', '.majorTicks', [1], tick);
        applyGridStyles(majorTicks, griddata);
    }

    if (renderGrid && showMinorGrid && gridSize) {
        // apply styles to minor grid
        // apply the transform
        minorGrids.attr('transform', d => transform(position(d)));
        applyGridStyles(minorGrids, minorGridData);
    }

    var hierarchy = [ticklabeldata, chart.systemConf.chart, chart.systemConf.canvas, $ZCG],
        fSize = utils.getVal(getValByPriority(hierarchy, 'fontSize')),
        fFamily = getValByPriority(hierarchy, 'fontFamily'),
        fWeight = getValByPriority(hierarchy, 'fontWeight'),
        fStyle = getValByPriority(hierarchy, 'fontStyle'),
        fShadow = getValByPriority(hierarchy, 'textShadow'),
        fColor = getValByPriority(hierarchy, 'fontColor'),
        txtDec = getValByPriority(hierarchy, 'textDecoration'),
        strokeColor = getValByPriority(hierarchy, 'strokeColor');
    // #ZC2801 subTickSize
    var rotationInfo = getTickRotateInfo(orient, rotation, null, null, Math.max(tickmarkdata.size, subTickSize), tickPadding, tickObj);

    ticklabel = ticklabel.merge(tickEnter.append('text')); //TODO Check appending ticklabels

    const ticklabelPosition = function(d, attr, val, callback) {
        if (callback) {
            let customTickTransform = invokeFunction(callback, d, scale1, chart);
            val += attr == 'x' ? customTickTransform[0] : customTickTransform[1];
        }
        return val;
    };
    if (renderTicklabel) {
        ticklabel
            .attr('class', 'ticklabel')
            .style('font-size', `${fSize}px`)
            .style('font-family', fFamily)
            .style('font-weight', fWeight)
            .style('font-style', fStyle)
            .style('text-shadow', fShadow)
            .style('text-decoration', txtDec)
            .style('fill', fColor)
            .style('text-anchor', rotationInfo.textAnchor)
            .style('cursor', cursor)
            .style(x, l * spacing)
            .attr(x, d => {
                return ticklabelPosition(d, x, l * spacing, tickTransform);
            })
            .attr(y, d => {
                return ticklabelPosition(d, y, 0, tickTransform);
            })
            .attr('dy', function(d, i) {
                return tickObj && tickObj.mode == 'zigzag' && i % 2 && rotationInfo.dy1 ? rotationInfo.dy1 : rotationInfo.dy;
            })
            .attr('transform', d => {
                let t = [0, 0];
                if (tickTransform) {
                    t = invokeFunction(tickTransform, d, scale1, chart);
                }
                return `rotate(${rotationInfo.rotate[0]} ${rotationInfo.rotate[1] + t[0]} ${rotationInfo.rotate[2] + t[1]})`;
            });
        updateMinmaxlabelPosition(ticklabel, orient, rotationInfo, ticklabeldata, data, scale1, chart);
        if (axis.ticks.isHierarchialLevelType()) {
            ticklabel.call(centeringTicklabel, scale1, orient);
        }
    }
    //HANDLE OVERFLOW MISMATCH occur if transition applied to tick transform.
    //Need to handle overlap by data and scale, not by posiiton
    if(chart.racing && chart.racing.syncTransition){
        tick.transition(transition_).attr('transform', function(d) {
            return transform(position(d));
        });
    }else{
        tick.attr('transform', function(d) {
            return transform(position(d));
        });
    }

    if (context !== selection) {
        // path = path.transition(context);
        tick = tick.transition(context);
        minorTicks = minorTicks.transition(context);
        ticklabel = ticklabel.transition(context);
        minorGrids = minorGrids.transition(context);
        subTicks = subTicks.transition(context);

        // TODO AARTHI DOUBT
        tickExit = tickExit
            .transition(transition_)
            .attr('opacity', epsilon)
            .attr('transform', function(d) {
                return isFinite((d = position(d))) ? transform(d) : this.getAttribute('transform');
            });

        tickEnter.attr('opacity', epsilon).attr('transform', function(d) {
            var p = this.parentNode.__axis;
            return transform(p && isFinite((p = p(d))) ? position(d) : position(d));
        });
    }

    tickExit.remove();

    path.attr(
        'd',
        orient === 'left' || orient == 'right'
            ? `M${k * (tickSizeOuter + tickmarkStart)},${range0}${h}0.5V${range1}H${k * (tickSizeOuter + tickmarkStart)}`
            : `M${range0},${k * tickSizeOuter}V0.5H${range1}V${k * tickSizeOuter}`
    );

    tick.attr('opacity', 1).attr('transform', function(d) {
        return transform(position(d));
    });
    if (tickObj && tickObj.isInline && (defined(plotBorderData) && plotBorderData.show === true)) {
        var isLeftOrRight = orient == 'left' || orient == 'right',
            translateIndex = isLeftOrRight ? 1 : 0,
            pos = isLeftOrRight ? fSize * parseFloat(rotationInfo.dy) : 0,
            margin = isLeftOrRight
                ? [0, chart.plotarea.height]
                : [chart.plotarea.left + chart.plotarea.left * 0.05, chart.plotarea.width + chart.plotarea.width * 0.05];
        tick.style('display', function() {
            var translate = getSvgTransform(this).translate[translateIndex];
            return isWithinRange(pos + translate, margin) ? '' : 'none';
        });
    }
    minorTicks.attr(`${x}2`, k * tickmarkStart + m * tickSizeInner);

    selection
        .filter(entering)
        .attr('fill', 'none')
        .attr('font-size', 10)
        .attr('font-family', 'sans-serif')
        .attr('text-anchor', orient === 'right' ? 'start' : orient === 'left' ? 'end' : 'middle');

    selection.each(function() {
        this.__axis = position;
    });

    // ticklabel
    //     .attr(x, k * spacing);
    if (renderTicklabel) {
        ticklabel.each(format);

        if (ticklabeldata.tooltip === true) {
            ticklabel.each(function(d) {
                addSVGtitleEle(d3_select(this), this.formattedVal);
            });
        }
    }
}
