// @zohocharts/zc/src/axis/dateinterval.js
//$Id$
//dateinterval.js
import { pick, defined } from '../utils/zcutils';
import { mathCeil, mathFloor } from '../core/zc-constants';
import { TIME_DATATYPE } from '../core/datatype-constants';
import {
    d3_timeSunday,
    d3_timeMonday,
    d3_timeTuesday,
    d3_timeWednesday,
    d3_timeThursday,
    d3_timeFriday,
    d3_timeSaturday,
    d3_timeSeconds,
    d3_timeMinutes,
    d3_timeHours,
    d3_timeDays,
    d3_timeWeeks,
    d3_timeMonths,
    d3_timeYears
} from '../core/d3-constants';
import { getFormattedValue } from '../formatter/formatter';

export var date_utils = function(formatConf) {
    formatConf = pick(formatConf, {}); // #ZC1187
    this.startWeekDay = defined(formatConf.startWeekDay) ? formatConf.startWeekDay : 1;
    this.startMonth = defined(formatConf.startMonth) ? formatConf.startMonth : 0;
};

date_utils.MILLISEC = 0;
date_utils.SECOND = 1;
date_utils.MINUTE = 2;
date_utils.HOUR = 3;
date_utils.DATE = 4;
date_utils.MONTH = 5;

date_utils.SECOND_TYPE = 'ABSSECOND'; //Seconds wise
date_utils.MINUTE_TYPE = 'ABSMINUTE'; //Minutes wise
date_utils.HOUR_TYPE = 'ABSHOUR'; //Hours wise
date_utils.DAY_TYPE = 'ABSDAY'; //Weekday wise(sunday,monday..)
date_utils.DATETIME_TYPE = 'DATETIME'; //Date with time wise(18 Sep 2011 01:10:40)
date_utils.DATE_TYPE = 'DATE'; //Date wise(18 Sep 2011)
date_utils.ABSWEEK_TYPE = 'ABSWEEK'; //Week number wise
date_utils.ABSMONTH_TYPE = 'ABSMONTH'; //Month wise
date_utils.ABSQUARTER_TYPE = 'ABSQUARTER'; //Quarter wise,four quarters for each year(Q1,Q2,Q3,Q4)
date_utils.ABSYEAR_TYPE = 'YEAR'; //Year wise

date_utils.D3_WEEKDAY_FNS = [
    d3_timeSunday,
    d3_timeMonday,
    d3_timeTuesday,
    d3_timeWednesday,
    d3_timeThursday,
    d3_timeFriday,
    d3_timeSaturday
];
date_utils.DAY_MILLIS = 24 * 60 * 60 * 1000;

date_utils.prototype.setZero = function(dateVal, type) {
    var oldVal = 0;
    for (var i = 0; i <= type; i++) {
        switch (i) {
            case date_utils.MILLISEC:
                {
                    oldVal += this.getMilliseconds(dateVal);
                    this.setMilliseconds(dateVal, 0);
                }
                break;
            case date_utils.SECOND:
                {
                    oldVal += this.getSeconds(dateVal);
                    this.setSeconds(dateVal, 0);
                }
                break;
            case date_utils.MINUTE:
                {
                    oldVal += this.getMinutes(dateVal);
                    this.setMinutes(dateVal, 0);
                }
                break;
            case date_utils.HOUR:
                {
                    oldVal += this.getHours(dateVal);
                    this.setHours(dateVal, 0);
                }
                break;
            case date_utils.DATE:
                {
                    oldVal += this.getDate(dateVal) - 1;
                    this.setDate(dateVal, 1);
                }
                break;

            case date_utils.MONTH:
                {
                    oldVal += this.getMonth(dateVal);
                    this.setMonth(dateVal, this.startMonth); // set custom startMonth
                }
                break;
            default:
                {
                    throw `unknown type 333: ${type}`;
                }
                break;
        }
    }
    return oldVal;
};

date_utils.prototype.getBaseValue = function(dateVal, type, isCeil) {
    switch (type) {
        case date_utils.SECOND_TYPE:
        case date_utils.DATETIME_TYPE:
            {
                var oldVal = this.setZero(dateVal, date_utils.MILLISEC);
                if (isCeil && oldVal > 0) {
                    this.setSeconds(dateVal, this.getSeconds(dateVal) + 1);
                }
            }
            break;
        case date_utils.MINUTE_TYPE:
            {
                var oldVal = this.setZero(dateVal, date_utils.SECOND);
                if (isCeil && oldVal > 0) {
                    this.setMinutes(dateVal, this.getMinutes(dateVal) + 1);
                }
            }
            break;
        case date_utils.HOUR_TYPE:
            {
                var oldVal = this.setZero(dateVal, date_utils.MINUTE);
                if (isCeil && oldVal > 0) {
                    this.setHours(dateVal, this.getHours(dateVal) + 1);
                }
            }
            break;
        case date_utils.DAY_TYPE:
        case date_utils.DATE_TYPE:
            {
                var oldVal = this.setZero(dateVal, date_utils.HOUR);
                if (isCeil && oldVal > 0) {
                    this.setDate(dateVal, this.getDate(dateVal) + 1);
                }
            }
            break;

        case date_utils.ABSWEEK_TYPE:
            {
                var oldVal = this.setZero(dateVal, date_utils.SECOND);
                var startWeekDay = this.startWeekDay;
                if (isCeil) {
                    if (oldVal > 0) {
                        this.setDate(dateVal, this.getDate(dateVal) + 1);
                    }
                    if (startWeekDay < this.getDay(dateVal)) {
                        this.setDate(dateVal, this.getDate(dateVal) + (7 - this.getDay(dateVal) + startWeekDay));
                    } else {
                        this.setDate(dateVal, this.getDate(dateVal) + (startWeekDay - this.getDay(dateVal)));
                    }
                } else {
                    if (startWeekDay > this.getDay(dateVal)) {
                        this.setDate(dateVal, this.getDate(dateVal) - (7 - startWeekDay + this.getDay(dateVal)));
                    } else {
                        this.setDate(dateVal, this.getDate(dateVal) - (this.getDay(dateVal) - startWeekDay));
                    }
                }
            }
            break;

        case date_utils.ABSQUARTER_TYPE:
        case date_utils.ABSMONTH_TYPE:
            {
                var oldVal = this.setZero(dateVal, date_utils.DATE);
                if (isCeil && oldVal > 0) {
                    this.setMonth(dateVal, this.getMonth(dateVal) + 1);
                }
            }
            break;

        case date_utils.ABSYEAR_TYPE:
            {
                var oldVal = this.setZero(dateVal, date_utils.MONTH);
                if (isCeil && oldVal > this.startMonth) {
                    //set base value based on custom start month not by default 0
                    this.setFullYear(dateVal, this.getFullYear(dateVal) + 1);
                }
            }
            break;
        default: {
            throw `unknow type 222: ${type}`;
        }
    }
};

date_utils.prototype.getTicksList = function(t0, t1, dt, type) {
    var times = new Array();
    var t1Time = t1.getTime();
    var time = t0.getTime();
    var t = new Date(time);

    while (time <= t1Time) {
        times.push(t);
        t = new Date(time);
        switch (type) {
            case date_utils.SECOND_TYPE:
            case date_utils.DATETIME_TYPE:
                {
                    t.setTime(t.getTime() + mathFloor(dt) * 1000);
                }

                break;

            case date_utils.MINUTE_TYPE:
                {
                    t.setTime(t.getTime() + dt * 60000);
                }
                break;
            case date_utils.HOUR_TYPE:
                {
                    t.setTime(t.getTime() + dt * 3600000);
                }
                break;

            case date_utils.DAY_TYPE:
            case date_utils.DATE_TYPE:
                {
                    t.setTime(t.getTime() + dt * date_utils.DAY_MILLIS);
                }
                break;
            case date_utils.ABSWEEK_TYPE:
                {
                    this.setDate(t, this.getDate(t) + dt);
                }
                break;

                break;
            case date_utils.ABSMONTH_TYPE:
            case date_utils.ABSQUARTER_TYPE:
                {
                    this.setMonth(t, this.getMonth(t) + dt);
                }
                break;

            case date_utils.ABSYEAR_TYPE:
                {
                    this.setFullYear(t, this.getFullYear(t) + dt);
                }
                break;
        }
        time = t.getTime();
    }
    return times;
};

date_utils.prototype.range = function(minDate, maxDate, tickCount, type, tickInterval) {
    minDate = new Date(minDate.getTime());
    maxDate = new Date(maxDate.getTime());

    this.getBaseValue(minDate, type, false);
    this.getBaseValue(maxDate, type, true);
    switch (type) {
        case date_utils.SECOND_TYPE:

        case date_utils.DATETIME_TYPE:
            {
                // var interval = Math.ceil((maxDate.getTime() - minDate.getTime()) / (tickCount * 1000));
                var minInterval = 1;
                var noOfSec = mathCeil((maxDate.getTime() - minDate.getTime()) / 1000);
                var interval = mathCeil(noOfSec / tickCount) || minInterval;
                interval = getInterval(interval, tickInterval);
                return {
                    TYPE: d3_timeSeconds,
                    INTERVAL: interval,
                    TICKS: this.getTicksList(minDate, maxDate, interval, type)
                };
            }
            break;

        case date_utils.MINUTE_TYPE:
            {
                var minInterval = 1;
                var interval = mathCeil((maxDate.getTime() - minDate.getTime()) / (tickCount * 60000)) || minInterval;
                interval = getInterval(interval, tickInterval);
                return {
                    // TODO check for using  d3_timeHours instead of d3_timeMinutes.
                    TYPE: d3_timeHours,
                    INTERVAL: interval,
                    TICKS: this.getTicksList(minDate, maxDate, interval, type)
                };
            }
            break;
        case date_utils.HOUR_TYPE:
            {
                var minInterval = 1;
                var interval = mathCeil((maxDate.getTime() - minDate.getTime()) / (tickCount * 3600000)) || minInterval;
                interval = getInterval(interval, tickInterval);
                return {
                    TYPE: d3_timeHours,
                    INTERVAL: interval,
                    TICKS: this.getTicksList(minDate, maxDate, interval, type)
                };
            }
            break;
        case date_utils.DAY_TYPE:
        case date_utils.DATE_TYPE:
            {
                var minInterval = 1;
                var interval = mathCeil((maxDate.getTime() - minDate.getTime()) / (tickCount * date_utils.DAY_MILLIS)) || minInterval;
                interval = getInterval(interval, tickInterval);
                return {
                    TYPE: d3_timeDays,
                    INTERVAL: interval,
                    TICKS: this.getTicksList(minDate, maxDate, interval, type)
                };
            }
            break;

        case date_utils.ABSWEEK_TYPE:
            {
                var minInterval = 1;
                var interval_week = mathCeil((maxDate.getTime() - minDate.getTime()) / (tickCount * date_utils.DAY_MILLIS * 7));
                interval_week = getInterval(interval_week, tickInterval);
                var interval = mathCeil(interval_week) * 7 || minInterval; //interval in weekdays
                return {
                    TYPE: d3_timeWeeks,
                    INTERVAL: interval,
                    TICKS: this.getTicksList(minDate, maxDate, interval, type)
                };
            }
            break;

        case date_utils.ABSMONTH_TYPE:
            {
                var minInterval = 1;
                var interval = mathCeil(this.getMonthsDiff(minDate, maxDate) / tickCount) || minInterval;
                interval = getInterval(interval, tickInterval);
                return {
                    TYPE: d3_timeMonths,
                    INTERVAL: interval,
                    TICKS: this.getTicksList(minDate, maxDate, interval, type)
                };
            }
            break;
        case date_utils.ABSQUARTER_TYPE:
            {
                var minInterval = 1;
                var interval_month = this.getMonthsDiff(minDate, maxDate) / tickCount || minInterval;
                var quarter = mathCeil(interval_month / 3);
                quarter = getInterval(quarter, tickInterval);
                var interval = quarter * 3;
                return {
                    TYPE: d3_timeMonths,
                    INTERVAL: interval,
                    TICKS: this.getTicksList(minDate, maxDate, interval, type)
                };
            }
            break;

        case date_utils.ABSYEAR_TYPE:
            {
                var minInterval = 1;
                var interval = mathCeil((maxDate.getFullYear() - minDate.getFullYear()) / tickCount) || minInterval;
                interval = getInterval(interval, tickInterval);
                return {
                    TYPE: d3_timeYears,
                    INTERVAL: interval,
                    TICKS: this.getTicksList(minDate, maxDate, interval, type)
                };
            }
            break;

        default: {
            throw `UNKNOWN TYPE 111: ${type}`;
        }
    }

    function getInterval(actualInterval, userInterval) {
        return userInterval != null && userInterval >= actualInterval ? userInterval : actualInterval;
    }
};

date_utils.MAX_DATE = new Date(2008, 8, 29, 23, 23, 23, 999);
date_utils.prototype.getMaxDate = function() {
    return date_utils.MAX_DATE;
};

date_utils.prototype.getMonthsDiff = function(minDate, maxDate) {
    return (this.getFullYear(maxDate) - this.getFullYear(minDate)) * 12 + (this.getMonth(maxDate) - this.getMonth(minDate));
};

date_utils.prototype.getYearsDiff = function(minDate, maxDate) {
    return (this.getFullYear(maxDate) - this.getFullYear(minDate)) * 12 + (this.getMonth(maxDate) - this.getMonth(minDate));
};

date_utils.prototype.getFullYear = function(date) {
    return this.useUTC != null && this.useUTC ? date.getUTCFullYear() : date.getFullYear();
};
date_utils.prototype.getMonth = function(date) {
    return this.useUTC != null && this.useUTC ? date.getUTCMonth() : date.getMonth();
};
date_utils.prototype.getDay = function(date) {
    return this.useUTC != null && this.useUTC ? date.getUTCDay() : date.getDay();
};
date_utils.prototype.getDate = function(date) {
    return this.useUTC != null && this.useUTC ? date.getUTCDate() : date.getDate();
};
date_utils.prototype.getHours = function(date) {
    return this.useUTC != null && this.useUTC ? date.getUTCHours() : date.getHours();
};
date_utils.prototype.getMinutes = function(date) {
    return this.useUTC != null && this.useUTC ? date.getUTCMinutes() : date.getMinutes();
};
date_utils.prototype.getSeconds = function(date) {
    return this.useUTC != null && this.useUTC ? date.getUTCSeconds() : date.getSeconds();
};
date_utils.prototype.getMilliseconds = function(date) {
    return this.useUTC != null && this.useUTC ? date.getUTCMilliseconds() : date.getMilliseconds();
};

date_utils.prototype.setFullYear = function(date, val) {
    return this.useUTC != null && this.useUTC ? date.setUTCFullYear(val) : date.setFullYear(val);
};
date_utils.prototype.setMonth = function(date, val) {
    return this.useUTC != null && this.useUTC ? date.setUTCMonth(val) : date.setMonth(val);
};
date_utils.prototype.setDate = function(date, val) {
    return this.useUTC != null && this.useUTC ? date.setUTCDate(val) : date.setDate(val);
};
date_utils.prototype.setHours = function(date, val) {
    return this.useUTC != null && this.useUTC ? date.setUTCHours(val) : date.setHours(val);
};
date_utils.prototype.setMinutes = function(date, val) {
    return this.useUTC != null && this.useUTC ? date.setUTCMinutes(val) : date.setMinutes(val);
};
date_utils.prototype.setSeconds = function(date, val) {
    return this.useUTC != null && this.useUTC ? date.setUTCSeconds(val) : date.setSeconds(val);
};
date_utils.prototype.setMilliseconds = function(date, val) {
    return this.useUTC != null && this.useUTC ? date.setUTCMilliseconds(val) : date.setMilliseconds(val);
};
