// @zohocharts/zc/src/axis/hierarchy-axis-utils.js
import { mergeJSON } from '../utils/zcutils';

export function getLeafAxisOrder(axisOrder, clusterCount, axisType, chartObj) {
    if (clusterCount == null) {
        let hierarchical = chartObj.dataObject.hierarchical,
            tag = axisType == 'y' ? 'row' : 'column',
            nodeList = hierarchical.pivotLayout.getRootNodeList(tag);
        clusterCount = nodeList.getClustersCount();
    }

    return axisOrder % clusterCount;
}

export function getAxisForLeaf(axisType, axisOrder, chartObj) {
    return chartObj.axes[axisType].leaves[axisOrder];
}

export function getAxisForBranch(axisType, axisOrder, chartObj) {
    return chartObj.axes[axisType].branches[axisOrder];
}

export function getAxisDataIndex(axisType, axisOrder, chartObj) {
    let hierarchical = chartObj.dataObject.hierarchical,
        isCategoryHierarchy = chartObj.dataObject.categoryHierarchy,
        tag = axisType == 'y' ? 'row' : 'column';
    if (hierarchical) {
        let nodeList = hierarchical.pivotLayout.getRootNodeList(tag);
        if (nodeList.getCount() || nodeList.hasMultiCluster()) {
            let clustersCount = nodeList.getClustersCount(),
                leavesCount = nodeList.getLeavesCount(),
                isLeafAxisType = isLeafAxis(axisType, axisOrder, chartObj),
                leafAxisOrder = getLeafAxisOrder(axisOrder, clustersCount);
            if (isLeafAxisType) {
                return leafAxisOrder;
            } else {
                return 0;
            }
        }
        return axisOrder;
    }
    return axisOrder;
}

export function isCategoryHierarchical(chartObj) {
    return chartObj.dataObject && chartObj.dataObject.hierarchical && chartObj.dataObject.hierarchical.categoryTree;
}

export function hasBranchAxis(axisType, chartObj) {
    let tag = axisType == 'x' ? 'column' : 'row',
        nodeList = chartObj.dataObject.hierarchical.pivotLayout.getRootNodeList(tag);

    return nodeList.getCount();
}

export function isBranchAxis(axisType, axisOrder, chartObj) {
    let tag = axisType == 'x' ? 'column' : 'row',
        nodeList = chartObj.dataObject.hierarchical.pivotLayout.getRootNodeList(tag),
        leavesCount = nodeList.getLeavesCount();

    return axisOrder > leavesCount - 1;
}

export function isLeafAxis(axisType, axisOrder, chartObj) {
    let tag = axisType == 'x' ? 'column' : 'row',
        nodeList = chartObj.dataObject.hierarchical.pivotLayout.getRootNodeList(tag),
        leavesCount = nodeList.getLeavesCount();

    return axisOrder < leavesCount;
}

export function getBranchDepth(axisType, axisOrder, chartObj) {
    let tag = axisType == 'x' ? 'column' : 'row',
        nodeList = chartObj.dataObject.hierarchical.pivotLayout.getRootNodeList(tag),
        leavesCount = nodeList.getLeavesCount();

    if (leavesCount > 1) {
        return axisOrder - leavesCount;
    }
    return axisOrder - 1;
}
export function isRootBranchAxis(axisType, axisOrder, chartObj) {
    let axesCount = chartObj.getAxesCount(axisType);

    return axisOrder == axesCount - 1;
}

export function showAxisLabelForTree(axisType, axisOrder, chartObj) {
    if (axisType == 'x') {
        return isRootBranchAxis(axisType, axisOrder, chartObj);
    } else {
        return isLeafAxis(axisType, axisOrder, chartObj) || isRootBranchAxis(axisType, axisOrder, chartObj);
    }
}

export function getAxisAreaDim(axisType, axisOrder, orient, chartObj, plotarea) {
    var hierarchical = chartObj.dataObject.hierarchical;
    if (hierarchical && hierarchical.pivot) {
        if (axisType == 'y' && isLeafAxis(axisType, axisOrder, chartObj)) {
            let tag = axisType == 'x' ? 'column' : 'row',
                nodeList = chartObj.dataObject.hierarchical.pivotLayout.getRootNodeList(tag),
                dimension = orient == 'left' || orient == 'right' ? 'height' : 'width',
                obj = {};
            obj[dimension] = (nodeList.leafSize * axisOrder + nodeList.leafSize / 2) * 2; //To calculate the relative position to locate axislabel
            return obj;
        }
        return plotarea;
    } else {
        return plotarea;
    }
}

export function mergeAxisOptionsForBranch(parent, level) {
    var paramsToMerge = ['axisline', 'grid', 'tickmark', 'label', 'ticklabel'];
    paramsToMerge.map(d => {
        mergeJSON(level[d], parent[d], false);
    });
    return level;
}

export function isFacet(chartObj) {
    let dataObject = chartObj.dataObject;
    if (!dataObject) {
        return false;
    }

    let { hierarchical } = dataObject;
    return hierarchical && hierarchical.pivot;
}
