// @zohocharts/zc/src/axis/log-tick-labels.js
import { arrayIncludes, calculatePercentage, defined, pick } from '../utils/zcutils';
import { filterCategories, getLinearTicksByCount, getTickWidth, nearestTickInterval } from './tick-labels';
import { d3_ascending, d3_extent, d3_range, d3_set } from '../core/d3-constants';
import { mathAbs, mathCeil, mathFloor, mathMax, mathMin, mathPow, UNDEFINED } from '../core/zc-constants';
import { getTextBound } from './textBound';
import { fixFloatingPoint, getMagnitude, logArgument, logBase } from '../utils/log-utils';
import { Registry } from '../core/zc-registry';
import { Console, scale } from '../core/main-zc-init';

const logTicks = scale.log.ticks = {};
/**
 *
 * @type {number}
 */
const INTERVAL_FACTOR = 1.5;

/**
 *
 * @param domain
 * @param axisSize
 * @param dataMinMax
 * @param usePadding
 * @param axis
 * @param scale
 * @param margin
 * @param styles
 * @param format
 * @param orient
 * @param rotation
 * @param adjustToBase
 * @param addMinMaxLog
 * @param hasFixedPosition
 * @param dataDomain
 * @return {{minorTickValues: [], tickValues: *}|{logDomain: *, minorTickValues: [], tickValues: *}}
 */

export const getLogTicks = logTicks.getLogTicks = function getLogTicks(
    domain,
    axisSize,
    dataMinMax,
    usePadding,
    axis,
    scale,
    margin,
    styles,
    format,
    orient,
    rotation,
    adjustToBase,
    addMinMaxLog,
    hasFixedPosition,
    dataDomain
) {
    let { minRange, maxRange, interval: userTickInterval, logBase: base = 10, outerPadding: padding, tickCount: userTickCount } = axis,
        minorGridOptions = axis.minorGrid || {},
        subTickMarkOptions = axis.subtickmark || {},
        showMinorGrid = minorGridOptions.show === true,
        allowNonPositiveTicks = axis.allowNonPositiveTicks !== false,
        minorGridCount = minorGridOptions.count,
        minorGridInterval = minorGridOptions.interval,
        // minor grid property has priority
        subTickCount = showMinorGrid ? pick(minorGridCount, subTickMarkOptions.count) : subTickMarkOptions.count,
        subTickInterval = showMinorGrid ? pick(minorGridInterval, subTickMarkOptions.interval) : subTickMarkOptions.interval,
        calculateMinorTicks = showMinorGrid || subTickCount > 0 || defined(subTickInterval),
        exponentToNumeric = exp => fixFloatingPoint(logArgument(exp, base, allowNonPositiveTicks), 12);

    let [min, max] = domain,
        forceUserTickCount,
        [dataMin, dataMax] = dataMinMax,
        minPadding = usePadding && !defined(minRange) ? pick(padding, 0.05) : 0,
        maxPadding = usePadding && !defined(maxRange) ? pick(padding, 0.05) : 0,
        tickCount = pick(userTickCount, axisSize / 70),
        minorTickValues = [],
        args,
        tickValues,
        logDomain,
        isSingleTick = min === max;

    // tick count has priority
    // either interval or count
    if (defined(userTickCount) || !userTickInterval) {
        userTickInterval = UNDEFINED;
    }

    if (userTickCount < 4) {
        forceUserTickCount = userTickCount;
        tickCount = userTickCount = 5;
    }

    // get the log of the min max
    let logMin = logBase(min, base, allowNonPositiveTicks),
        logMax = logBase(max, base, allowNonPositiveTicks),
        diff = logMax - logMin || 1;

    // add the padding to the log value
    logMin -= diff * minPadding;
    logMax += diff * maxPadding;

    // Single Tick = test cases = FEATURES_LOGAXIS_45 CXZA & LOGAXIS_19
    if (logMin === logMax || (dataDomain[0] === 0 && dataDomain[1] === 0)) {
        return { tickValues: dataMinMax, minorTickValues: [] };
    }

    if (defined(adjustToBase)) {
        // Avoid unnecessary ticks because of axis padding
        // EG. for data [0.01] with 0.05% padding it would make -0.03993004518863 and when it is converted to linear number
        // it may be nearest base rounded to -1 which is negative -10
        if ((logMin < 0 || logMax < 0) && allowNonPositiveTicks) {
            if (logMin < adjustToBase && dataMin >= adjustToBase) {
                logMin = adjustToBase;
            } else if (logMax > adjustToBase && dataMax <= adjustToBase) {
                logMax = adjustToBase;
            }
        }
    }

    // construct the arguments
    args = {
        numberMin: min,
        numberMax: max,
        userTickCount,
        tickCount,
        logMin,
        logMax,
        forceUserTickCount,
        base,
        allowNonPositiveTicks,
        axisSize,
        tickMultiples: axis.tickMultiples
    };

    // Math.pow(10, 309) is Infinity.
    if (isInfinity(mathPow(base || 10, userTickInterval))) {
        Console.warn(`Log axis interval ${userTickInterval} is too large. Switching to the default interval.`);
        userTickInterval = null;
    }

    let interval = pick(userTickInterval, (logMax - logMin) / Math.max(tickCount - 1, 1)),
        tickValuesCopy;
    tickValues = getTickValues(args, interval, styles, orient, scale, format, margin, rotation);
    tickValuesCopy = tickValues.slice(0);
    logDomain = d3_extent(tickValues);

    // #ZC2835 user ranges
    if (defined(minRange)) {
        logDomain[0] = minRange;
        tickValues = tickValues.filter(value => value >= min);
    }
    if (defined(maxRange)) {
        logDomain[1] = maxRange;
        tickValues = tickValues.filter(value => value <= max);
    }

    let tickValuesWithMinMax = addMinMaxLog(tickValues),
        fixSpace = true;

    if (tickValuesWithMinMax.length !== tickValues.length) {
        tickValues = tickValuesWithMinMax;
        fixSpace = false;
    }

    // get the padded min max
    min = exponentToNumeric(logMin);
    max = exponentToNumeric(logMax);

    // *************** SMART SPACING ******************
    if (
        !defined(userTickCount) &&
        !defined(forceUserTickCount) &&
        !defined(userTickInterval) &&
        fixSpace &&
        !isSingleTick &&
        tickValues.length > 3
    ) {
        let [tickMin, tickMax] = logDomain,
            magMin = getMagnitude(min, base, allowNonPositiveTicks),
            magMax = getMagnitude(max, base, allowNonPositiveTicks),
            spaceMin = min - mathAbs(magMin) / 10,
            spaceMax = max + mathAbs(magMax) / 10,
            thresholdLimit = 40,
            filterFinalTick = false,
            // actual screen positions
            pos = tickValues.map(scale),
            totalPointDistanceMax = pos[pos.length - 1] - pos[pos.length - 2],
            pointDistanceMax = scale(max),
            // relative distance from the point
            // let pointDistanceMaxLeft = pointDistanceMax - pos[pos.length - 2];
            pointDistanceMaxRight = pos[pos.length - 1] - pointDistanceMax,
            // calculate the distance from the left and right
            // let distanceLeftPercentMax = calculatePercentage(pointDistanceMaxLeft, totalPointDistanceMax);
            distanceRightPercentMax = calculatePercentage(pointDistanceMaxRight, totalPointDistanceMax),
            // min side
            totalPointDistanceMin = pos[1] - pos[0],
            pointDistanceMin = scale(min),
            // let pointDistanceMinRight = pos[1] - pointDistanceMin;
            pointDistanceMinLeft = pointDistanceMin - pos[0],
            // let pointDistanceMinLeft2 = pointDistanceMin - spaceMin;

            distanceLeftPercentMin = calculatePercentage(pointDistanceMinLeft, totalPointDistanceMin);

        if (hasFixedPosition) {
            // ZC2805 data domain
            logDomain = domain;
            tickValues = tickValues.filter(value => value >= dataDomain[0] && value <= dataDomain[1]);
        } else if (distanceRightPercentMax > thresholdLimit || distanceLeftPercentMin > thresholdLimit) {
            if (distanceRightPercentMax > thresholdLimit && spaceMax < tickMax) {
                logDomain[1] = spaceMax;
            }

            if (distanceLeftPercentMin > thresholdLimit && spaceMin > tickMin) {
                logDomain[0] = spaceMin;
            }
            tickValues = tickValues.filter(value => value >= logDomain[0] && value <= logDomain[1]);
        }
    } else if (defined(args.useDefaultLogDomain)) {
        if (tickValues.every(tick => tick > min)) {
            logDomain[0] = min;
        }

        if (tickValues.every(tick => tick < max)) {
            logDomain[1] = max;
        }
    }

    if (calculateMinorTicks && !isSingleTick) {
        // TODO un-comment this line to fix
        // minorTickValues = getMinorTickValues(args, tickValuesCopy, logDomain, subTickCount, subTickInterval);
        // TODO tick values copy
        minorTickValues = getMinorTickValues(args, tickValues, logDomain, subTickCount, subTickInterval);
    }

    return { tickValues, minorTickValues, logDomain };
}

/**
 *
 * @param args
 * @param interval
 * @param styles
 * @param orient
 * @param scale
 * @param format
 * @param margin
 * @param rotation
 * @param isSubTicks
 * @return {any}
 */
function getTickValues(args, interval, styles, orient, scale, format, margin, rotation, isSubTicks) {
    args.interval = interval;

    let { base, allowNonPositiveTicks, tickCount, axisSize } = args,
        exponentToNumeric = exp => fixFloatingPoint(logArgument(exp, base, allowNonPositiveTicks), 12);

    let tickExponents = getLogTickExponents(args, isSubTicks),
        tickValues = tickExponents.map(exponentToNumeric),
        logDomain = d3_extent(tickValues),
        formatted,
        textDimension = text => {
            let dimension = getTextBound(text, null, null, styles),
                width = getTickWidth('width', dimension, margin, rotation, styles.fontSize, orient),
                height = getTickWidth('height', dimension, margin, rotation, styles.fontSize, orient);

            return {
                width,
                height
            };
        };

    tickValues = d3_set(tickValues)
        .values()
        .map(Number)
        .sort(d3_ascending);

    if (!isSubTicks) {
        formatted = format(tickValues);

        // scale
        scale.domain(logDomain);
        if (tickValues.length > tickCount && rotation === 'auto') {
            rotation = 45;
        }

        let bounds = tickValues.map(function(tick, i) {
            let x = 0,
                y = 0;
            if (orient === 'top' || orient === 'bottom') {
                x = scale(tick);
            } else {
                y = scale(tick);
            }

            return { ...textDimension(formatted[i]), x, y, tick: formatted[i] };
        });

        if (rectOverlapsWithAnother(bounds)) {
            // when the tick overlaps with another tick, we double the interval and re-generate the values;
            return getTickValues(args, interval * INTERVAL_FACTOR, styles, orient, scale, format, margin);
        }
    }

    return tickValues;
}

Registry.setComponent('logTicks.getLogTickValues', getTickValues);

/**
 *
 * @param rect1
 * @param rect2
 * @return {boolean}
 */
function overlapsWith(rect1, rect2) {
    return (
        rect1.x < rect2.x + rect2.width &&
        rect1.x + rect1.width > rect2.x &&
        rect1.y < rect2.y + rect2.height &&
        rect1.height + rect1.y > rect2.y
    );
}

function rectOverlapsWithAnother(bounds) {
    let visibleTicks = [],
        overlaps = false;

    bounds.forEach(function(tick, i) {
        if (overlaps) {
            return;
        }
        if (!visibleTicks.length) {
            visibleTicks.push(tick);
        } else {
            if (!(overlaps = visibleTicks.some(visibleLabel => overlapsWith(tick, visibleLabel)))) {
                visibleTicks.push(tick);
            }
        }
    });
    return overlaps;
}

/**
 *
 * @param args
 * @param tickValues
 * @param logDomain
 * @param minorGridCount
 * @param minorGridInterval
 * @return {*[]}
 */
function getMinorTickValues(args, tickValues, logDomain, minorGridCount, minorGridInterval = 'auto') {
    let minorTickValues = [],
        [min, max] = logDomain,
        minorTickInterval = minorGridInterval === 'auto' ? args.interval / 5 : minorGridInterval;

    if (minorGridCount > 0) {
        tickValues.forEach((value, index, array) => {
            if (index) {
                let min = mathMin(array[index - 1], value),
                    max = mathMax(array[index - 1], value),
                    diff = max - min,
                    interval = diff / (minorGridCount + 1),
                    steps = d3_range(min, max, interval);

                minorTickValues = [...minorTickValues, ...steps];
            }
        });
    } else {
        let { base, allowNonPositiveTicks } = args,
            exponentToNumeric = exp => fixFloatingPoint(logArgument(exp, base, allowNonPositiveTicks), 12);

        args = { ...args };
        args.interval = minorTickInterval;
        args.isMinorGrid = true;

        tickValues.forEach((value, index) => {
            if (index) {
                args.logMin = logBase(tickValues[index - 1], base, allowNonPositiveTicks);
                args.logMax = logBase(tickValues[index], base, allowNonPositiveTicks);

                let steps = getLogTickExponents(args, true);
                minorTickValues = [...minorTickValues, ...steps];
            }
        });
        minorTickValues = minorTickValues.map(exponentToNumeric);
    }

    return minorTickValues.filter(d => !arrayIncludes(tickValues, d) && d >= min && d <= max);
}

Registry.setComponent('logTicks.getMinorTickValues', getMinorTickValues);

/**
 *
 * @param args
 * @param isSubTicks
 * @return {*[]}
 */
function getLogTickExponents(args, isSubTicks) {
    let tickValues,
        { logMin, logMax, tickCount, interval, base, allowNonPositiveTicks, userTickCount, isMinorGrid, tickMultiples } = args,
        isIntermediate = false;

    if (interval >= 0.5) {
        interval = Math.round(interval);
        tickValues = getNiceLinearTicks(logMin, logMax, interval);
    } else if (interval >= 0.1) {
        isIntermediate = true;
        tickValues = getIntermediateLogTicks(logMin, logMax, interval, base, allowNonPositiveTicks, isMinorGrid);
    } else {
        // #ZC2802
        let numberMin = logArgument(logMin, base, allowNonPositiveTicks),
            numberMax = logArgument(logMax, base, allowNonPositiveTicks),
            diffs = numberMax - numberMin,
            numericInterval = diffs / tickCount,
            // DEMO_AREA_3 - decimal places fix
            mag = numericInterval > 1 ? getMagnitude(numericInterval, base, allowNonPositiveTicks) : getMagnitude(numericInterval, base),
            multiples = tickMultiples || [1, 2, 2.5, 5, 10];

        if (userTickCount > 0) {
            tickValues = getLinearTicksByCount([numberMin, numberMax], tickCount);
            args.useDefaultLogDomain = true;
        } else {
            numericInterval = pick(args.autoSubTickInterval, nearestTickInterval(numericInterval, mag, multiples));
            tickValues = getNiceLinearTicks(numberMin, numberMax, numericInterval);
        }
        if(tickValues.length > 10000 && logMax - logMin < 0.1){
            tickValues = getLinearTicksByCount([numberMin, numberMax], tickCount);
            args.useDefaultLogDomain = true;
        }
        tickValues = tickValues.map(d => logBase(d, base, allowNonPositiveTicks));

        if (!isSubTicks) {
            args.autoSubTickInterval = numericInterval / 5;
        }
    }

    // fix the min and max
    args.interval = interval;
    args.logMin = tickValues[0];
    args.logMax = tickValues[tickValues.length - 1];

    if (!isSubTicks && args.userTickCount > 0 && tickValues.length !== args.userTickCount) {
        // fix the user tick count
        args.extent = args.extent || d3_extent(tickValues);
        tickValues = getTickValuesByUserTickCount(tickValues, args, isIntermediate);
        tickValues = [...tickValues, ...args.extent];
    }

    return tickValues;
}

Registry.setComponent('logTicks.getLogTickExponents', getLogTickExponents);
/**
 *
 * @param logMin
 * @param logMax
 * @param interval
 * @param base
 * @param allowNonPositiveTicks
 * @param isMinorGrid
 * @return {[]}
 */

function getIntermediateLogTicks(logMin, logMax, interval, base, allowNonPositiveTicks, isMinorGrid) {
    var roundedMin = mathFloor(logMin),
        roundedMax = mathCeil(logMax),
        intermediateFactors,
        pos,
        skip,
        tickValues = [],
        filter = false,
        ticks = [];

    if (interval > 0.3) {
        intermediateFactors = [1, 2, 4];
        // 0.2 equals five minor ticks per 1, 10, 100 etc
    } else if (interval > 0.15) {
        intermediateFactors = [1, 2, 4, 6, 8];
    } else {
        // 0.1 equals ten minor ticks per 1, 10, 100 etc
        intermediateFactors = [1, 2, 3, 4, 5, 6, 7, 8, 9];
    }

    for (let i = roundedMin; i <= roundedMax + 1 && !skip; i++) {
        for (let j = 0; j < intermediateFactors.length && !skip; j++) {
            let val,
                checkMax = true;

            // #ZC2866 $ i + 1 in case of negative
            if (isMinorGrid && allowNonPositiveTicks && i < 0) {
                val = fixFloatingPoint(logArgument(i + 1, base, allowNonPositiveTicks)) * intermediateFactors[j];
                checkMax = false;
            } else {
                val = fixFloatingPoint(logArgument(i, base, allowNonPositiveTicks)) * intermediateFactors[j];
            }

            pos = logBase(val, base, allowNonPositiveTicks);

            if (pos >= roundedMin && pos <= roundedMax) {
                tickValues.push(pos);
                ticks.push(val);
            }

            if (checkMax && pos >= logMax) {
                skip = true;
            }
        }
    }

    // fix less than base both side
    if (ticks.some(v => v >= 0 && v <= base)) {
        let positions = intermediateFactors.map(f => logBase(f, base, allowNonPositiveTicks));
        tickValues = tickValues.concat(positions);
        filter = true;
    }

    if (ticks.some(v => v <= 0 && v >= -base)) {
        let positions = intermediateFactors.map(f => -logBase(f, base, allowNonPositiveTicks));
        tickValues = tickValues.concat(positions);
        filter = true;
    }

    if (filter) {
        tickValues = tickValues.filter(v => v >= roundedMin && v <= roundedMax);
    }

    tickValues.sort(d3_ascending);

    let last = null;
    for (let i = 0; i < tickValues.length; i++) {
        let current = tickValues[i];
        if (current > logMin) {
            break;
        }
        last = current;
    }

    if (defined(last)) {
        tickValues = tickValues.splice(tickValues.lastIndexOf(last));
    }

    return tickValues;
}

Registry.setComponent('logTicks.getIntermediateLogTickExponents', getIntermediateLogTicks);
/**
 *
 * @param tickValues
 * @param args
 * @return {*}
 */
function getTickValuesByUserTickCount(tickValues, args, isIntermediate) {
    let { interval: tickInterval, tickCount, forceUserTickCount, logMin: min, logMax: max, base, allowNonPositiveTicks, extent } = args;
    var len,
        i,
        currentTickCount = tickValues && tickValues.length,
        minValue = 0;

    let getIntermediateValue = side => {
        let v = side === -1 ? tickValues[tickValues.length - 1] + tickInterval : tickValues[0] - tickInterval,
            val = logArgument(fixFloatingPoint(v), base, allowNonPositiveTicks),
            mag = getMagnitude(val, base, allowNonPositiveTicks),
            v2 = side === -1 ? Math.ceil(val / mag) * mag : Math.floor(val / mag) * mag;

        return logBase(fixFloatingPoint(v2), base, allowNonPositiveTicks);
    };

    if (currentTickCount < tickCount) {
        while (tickValues.length < tickCount) {
            if (tickValues.length % 2 || minValue === min) {
                // to the end
                if (isIntermediate) {
                    tickValues.push(getIntermediateValue(-1));
                } else {
                    tickValues.push(fixFloatingPoint(tickValues[tickValues.length - 1] + tickInterval));
                }
            } else {
                if (isIntermediate) {
                    tickValues.unshift(fixFloatingPoint(getIntermediateValue(1)));
                } else {
                    // to the start
                    tickValues.unshift(fixFloatingPoint(tickValues[0] - tickInterval));
                }
            }
        }
    } else if (currentTickCount > tickCount) {
        // double the interval and generate the ticks
        // args.interval *= INTERVAL_FACTOR;
        // tickValues = getLogTickExponents(args);

        tickValues = filterCategories(tickValues, tickCount);
    }

    if (defined(forceUserTickCount) && tickValues.length !== forceUserTickCount) {
        let [min, max] = d3_extent(tickValues),
            center = tickValues[Math.round(tickValues.length / 2)];

        if (forceUserTickCount === 2) {
            tickValues = [min, max];
        } else {
            tickValues = [min, center, max];
        }
        args.forceUserTickCount = null;
    }

    tickValues = tickValues.filter(value => value >= extent[0] && value <= extent[1]);
    return tickValues;
}

Registry.setComponent('logTicks.getTickValuesByUserTickCount', getTickValuesByUserTickCount);

/**
 *
 * @param min
 * @param max
 * @param interval
 * @return {*}
 */
function getNiceLinearTicks(min, max, interval) {
    let roundedLogMin = fixFloatingPoint(Math.floor(min / interval) * interval),
        roundedLogMax = fixFloatingPoint(Math.ceil(max / interval) * interval),
        // d3 inclusive the interval for last tick
        maxInclusiveCorrection = interval / 2;

    return d3_range(roundedLogMin, roundedLogMax + maxInclusiveCorrection, interval);
    // return getLinearTicksByInterval([roundedLogMin, roundedLogMax + maxInclusiveCorrection], interval);
}

Registry.setComponent('logTicks.getNiceLinearTicks', getNiceLinearTicks);

export function isInfinity(number) {
    return number === Infinity || number === -Infinity;
}
