// @zohocharts/zc/src/axis/scale.js
//$Id$
//scale.js
import { charttypenames, charttype, scale, chartcategory } from '../core/main-zc-init';
import {
    defined,
    getPropVal,
    isWithinRange,
    objectValues,
    pick,
    validateNullVal,
    isArray,
    invokeFunction,
    isUndefined,
    objectToArray
} from '../utils/zcutils';
import { colorscale_helpers } from './colorscale-helpers';
import { mathAbs, mathMin, UNDEFINED } from '../core/zc-constants';
import {
    d3_ascending,
    d3_descending,
    d3_range,
    d3_scaleLinear,
    d3_scaleOrdinal,
    d3_sum,
    d3_extent,
    d3_max,
    d3_scaleSymlog,
    d3_scaleSqrt,
    d3_map,
    d3_merge
} from '../core/d3-constants';
import { nonLinearOrdinalScale } from './nonLinearOrdinalScale';
import { isBarChartType, isHeatMapChartType } from '../events/events-utils';
import { getPlotSize } from '../plot/plotSize';
import { DataProcessor } from '../core/data-processor'; //Trash
import { NUMERIC_DATATYPE, ORDINAL_DATATYPE, TIME_DATATYPE } from '../core/datatype-constants';
import { getLeafAxisOrder, isFacet } from './hierarchy-axis-utils';
import { hasColorColumnInNonAxis } from '../utils/colorutils';
// import { getPolarRange, getPolarRangeConfig } from '../polaraxis/polaraxis-helpers';
import { bisectArray } from '../utils/array-utils';
import { BOTTOM_ALIGN, LEFT_ALIGN, TOP_ALIGN } from '../core/align-constants';
import { getAxisLayout, getAxisOrient } from './axisutils';
import { getGridList } from '../core/data-processor/pivot-layout-helpers';
import { LOG } from '../core/scale-type-constant';
import { getChartTypeByIndex, isCategoryOf } from '../core/charttype-helpers';

export * from './quantileScale';

import { expressionEvalRangeForLinear } from '../scale/evaluators/evalRangeNumericAxisParser';
import { Scale } from '../scale/scale-component';
// import { numericAxisParser } from '../scale-parser/axis/numeric-axis-parser';
import { LINEAR_SCALETYPE, POLAR_SCALETYPE, QUANTILE_SCALETYPE } from '../scale/scale-type-constants';
import { ScaleRegistry } from '../scale/scale-registry';
// import { ordinalAxisRangeParser } from '../scale-parser/axis/ordinal-axis-parser';
import { NULL } from '../core/zc-constants';
import { expressionEvalDomainForLinear } from '../scale/evaluators/evalDomainNumericAxisParser';
import { expressionEvalRangeForOrdinal } from '../scale/evaluators/evalRangeOrdinalAxisParser';
import { extendScaleParser } from '../parser/parser-utils';
import { getAxisViewModel, getAxisData } from './axisutils';
import { Registry } from '../core/zc-registry';
import { getSizeFromPoint } from '../core/data-processor/size/size-from-point';

scale.getScale = function(chartObj, dataObj, plotarea, orient, index, dataset, type, extent, noOuterPadding, applyZoom = true) {
    var scaleObj, newScaleObj, domain;
    var systemconf = chartObj.systemConf;
    var datatype = dataset.getDataType(type, index),
        axis_data = getAxisOptions(chartObj, type, index),
        hierarchical = dataObj.hierarchical;

    let isPolarScaleType = dataset.getScaleType?.(type, index) == 'polar';

    /*
     * If hasCustomMinRange or hasCustomMaxRange is present, the correction should not be applied to the scale domain
     * Test cases :- Test_535.json, Test-536.json
     */
    var hasCustomMinRange = defined(axis_data.minRange),
        hasCustomMaxRange = defined(axis_data.maxRange);

    var cType = charttypenames.get(getFirstAxisChartType(chartObj));
    var commonPlotOptions = systemconf.chart.plot.plotoptions[cType];
    var isPolar = cType === 'dial' || (cType === 'web' && type === 'x') || isPolarScaleType, //isPolarType(chartObj, type, index),
        isHeatmap = cType === 'heatmap',
        isBarChart = cType === 'bar',
        hasColorScale = type === 'clr' && colorscale_helpers.isQuantileColorScale(chartObj),
        columnInfo = dataset.getColumn(type, index),
        // [**MIGRATION**] zoom --> brush
        // #ZC2017
        zoomedValue = applyZoom ? getPropVal(systemconf, `chart.brush.${type}[${index}]`) : UNDEFINED,
        isLogScale = columnInfo.scaleType === LOG;

    const filteredIndices = isPolarScaleType && isBarChart ? chartObj.dataObject.filterStore.x() : null;

    var range = getRange(chartObj, type, orient, axis_data.reversed, index),
        padding = getPadding(),
        outerPadding = getOuterPadding();

    if (type === 'clr') {
        // domain = colorscale_helpers.getColorDomain(chartObj);
        // range = colorscale_helpers.getProcessedColorRange(range, domain, systemconf.legend, chartObj);

        let rangeOperators_clr, domainOperators_clr;

        domainOperators_clr = ScaleRegistry.get('color.domainParser')(type, index, systemconf.legend, chartObj);
        rangeOperators_clr = ScaleRegistry.get('color.rangeParser')(type, index, systemconf.legend, chartObj);
        var parser1 = {
            type: datatype,
            dataType: datatype,
            scaleType: datatype != ORDINAL_DATATYPE && axis_data.ranges && axis_data.ranges.length ? QUANTILE_SCALETYPE : datatype,
            field: columnInfo,
            reverse: axis_data.reversed,
            data: [],
            domain: {
                operators: domainOperators_clr,
                evaluator: ScaleRegistry.get('color.evalDomain')
            },
            range: {
                operators: rangeOperators_clr,
                evaluator: ScaleRegistry.get('color.evalRange')
            },
            unknown: 0,
            padding: padding,
            outerPadding: outerPadding
        };
        // if (chartObj.viewFlow && chartObj.viewFlow.scales[columnInfo.id]) {
        //     extendScaleParser(parser1, chartObj.viewFlow.scales[columnInfo.id].conf);
        // }
        if (chartObj.viewFlow) {
            let _scale = chartObj.viewFlow.getScaleByColumnName(columnInfo, type);
            if(_scale){
                extendScaleParser(parser1, _scale.conf, chartObj);
            }
            
        }
        // if (systemconf.viewParser) {
        //     if (systemconf.viewParser.scale && systemconf.viewParser.scale[columnInfo.name]) {
        //         extendScaleParser(parser1, systemconf.viewParser.scale[columnInfo.name]);
        //     }
        // }
        var newScaleComp1 = new Scale(type, index, parser1, chartObj);
        newScaleObj = newScaleComp1.getScale();
    }

    switch (datatype) {
        case 'numeric':
        case 'time':
            // if (extent) {
            //     domain = extent;
            // } else if (type === 'x') {
            //     domain = dataObj.getMinMaxX(0, applyZoom);
            // } else if (type === 'y') {
            //     domain = dataObj.getMinMaxY(index, outerPadding, UNDEFINED, applyZoom);
            // }
            // var scaleMode;
            //
            // if (type === 'clr' && axis_data.ranges && axis_data.ranges.length) {
            //     scaleMode = colorscale_helpers.quantileScale;
            // } else {
            //     scaleMode = getContinuousScale(dataset, type, index);
            // }
            //
            // var actualStart = domain[0],
            //     actualEnd = domain[1];
            // if (domain[0] == 'NoData' && domain[1] == 'NoData') {
            //     //for addorremoveaxis
            //     domain = [0, 0];
            // }
            // if (!zoomedValue && (extent || type == 'x')) {
            //     var diff = domain[1] - domain[0];
            //     var correction = diff == 0 ? chartObj.plotarea.width / 2 : diff * outerPadding;
            //     domain[0] = hasCustomMinRange ? domain[0] : domain[0] - correction;
            //     domain[1] = hasCustomMaxRange ? domain[1] : domain[1] + correction;
            // }
            //
            // scaleObj = scaleMode()
            //     .domain(domain)
            //     .range(range);
            //
            // // TODO CXZA_SINGLECOL_2
            //
            // // SYM LOG SCALE constant for log like spacing near 0 ticks. eg [0.001, 0.01, 0.1, 1, 10]
            // if (scaleObj.constant) {
            //     let min = d3_min(domain);
            //     if (min > 0 && min < pick(axis_data.logBase, 10) && axis_data.allowNonPositiveTicks === false) {
            //         scaleObj.constant(0.0001);
            //     } else {
            //         scaleObj.constant(1);
            //     }
            //     // #ZC2812
            //     if (actualStart === 'NoData' && actualEnd === 'NoData') {
            //         actualStart = actualEnd = 0;
            //     }
            // }
            //
            // var originalDomain = [actualStart, actualEnd],
            //     actualRange = originalDomain.map(scaleObj);
            //
            // scaleObj.actualRange = actualRange;
            // scaleObj.origRange = range;
            // scaleObj.originalDomain = originalDomain;
            // scaleObj.correction = correction;

            ////
            /*
                SCALE PARSER
            */
            if (type == 'x' || type == 'y') {
                var domainOperators = {},
                    rangeOperators = {};

                var orient = getAxisOrient(type, chartObj.axes.rotated),
                    minmaxCorrection = datatype == 'time' ? NULL : type == 'x' ? 0 : outerPadding,
                    scaleType = isPolar
                        ? POLAR_SCALETYPE
                        : axis_data.ranges && axis_data.ranges.length
                        ? QUANTILE_SCALETYPE
                        : columnInfo.scaleType,
                    polarConfig = Registry.getComponent('axis.polar.helpers.getPolarRangeConfig')(chartObj, type);

                rangeOperators.plotBound = {
                    dimension: orient == BOTTOM_ALIGN || orient == TOP_ALIGN ? 'width' : 'height',
                    customPlotBound: { width: chartObj.systemConf.chart.width, height: chartObj.systemConf.chart.height },
                    plotSize: chartObj.plotarea,
                    defaultPlotSize: chartObj.plotarea,
                    polarConfig,
                    color: chartObj.imagePallete || chartObj.seriesColor,
                    boundExtreme: 0
                };

                domainOperators = Registry.getComponent('axis.parser.numeric')(
                    type,
                    index,
                    chartObj,
                    outerPadding,
                    UNDEFINED,
                    applyZoom,
                    UNDEFINED,
                    minmaxCorrection,
                    extent
                );

                var parser = {
                    dataType: datatype,
                    orient: datatype == TIME_DATATYPE ? chartObj.axes.x.orient : orient,
                    type: datatype,
                    scaleType,
                    field: columnInfo,
                    reverse: axis_data.reversed,
                    data: [],
                    domain: {
                        operators: domainOperators,
                        evaluator: expressionEvalDomainForLinear
                    },
                    range: {
                        operators: rangeOperators,
                        evaluator: expressionEvalRangeForLinear
                    },
                    padding: padding,
                    outerPadding: outerPadding
                };
                //for log scale
                if (columnInfo.scaleType === LOG) {
                    parser.logBase = axis_data.logBase;
                }
                if (chartObj.viewFlow) {
                    let _scale = chartObj.viewFlow.getScaleByColumnName(columnInfo, type);
                    if(_scale){
                        if(chartObj.customVisual){
                            // let scale_customVisual_conf = simpleClone(_scale.conf);
                            //apply mandatory parser conf
                            // scale_customVisual_conf.dataType = scale_customVisual_conf.dataType || parser.dataType;
                            // scale_customVisual_conf.orient = scale_customVisual_conf.orient || parser.orient
                            // parser = _scale.conf;
                        }
                        extendScaleParser(parser, _scale.conf, chartObj);
                    }
                    
                }
                // if (systemconf.viewParser) {
                //     if (systemconf.viewParser.scale && systemconf.viewParser.scale[columnInfo.name]) {
                //         extendScaleParser(parser, systemconf.viewParser.scale[columnInfo.name]);
                //     }
                // }
                var newScaleComp = new Scale(type, index, parser, chartObj);
                newScaleObj = newScaleComp.getScale();

                if (type == 'y') {
                    //TODO Need to change this dependency from dataobject
                    dataObj.hasNegativeValue[index] = newScaleComp.config.hasNegativeValue && newScaleComp.config.hasNegativeValue[index];
                }
            }
            // else {
            //     newScaleObj = newScaleObj;
            // }
            getActualRange(newScaleObj);
            if (extent && datatype == NUMERIC_DATATYPE) {
                if (chartObj.scroll && chartObj.scroll.scrollDomain) {
                    chartObj.viewPortExtent = newScaleObj.domain();
                    chartObj.scroll.scrollDomain = chartObj.viewPortExtent;
                }
            }
            break;
        case 'ordinal':
            //args for customer will be fields not by encodes or chart types. they can give it as functions. also they can use operators
            // var parser = {
            //     type: 'ordinal',
            //     field: '',
            //     reverse: '',
            //     data: [],
            //     domain: {
            //         extremes: [
            //             {
            //                 data: [],
            //                 encode: ''
            //             }
            //         ],
            //         categories: {
            //             encode: ''
            //         },
            //         min: 'config.minRange',
            //         max: 'config.maxRange',
            //         ranges: 'config.ranges',
            //         stops: 'config.stops',
            //         base: 'config.base'
            //     },
            //     range: {
            //         plotBound: {
            //             dimension: 'width'
            //         },
            //         outerPadding: {}
            //     },
            //     padding: padding
            // };

            //YOU NEED TO AUTOMATE THIS PARSER AND VALUES PROPERLY;
            if (extent) {
                domain = extent;
            } else if (type !== 'clr') {
                var categories = dataObj.getCategories(type, index, applyZoom);
                filteredIndices?.forEach(_ => categories.splice(_, 1));
                domain = dataObj.getDomain(categories, type, true);
            }
            //
            // var d3_outerPadding = 0,
            //     origRange = [range[0], range[1]],
            //     outerPaddingPx;
            // //IMPORTANT: Padding is considered for single data. Refer /#/demoCharts/bullet1
            // padding = domain.length === 1 ? 0 : padding;
            //
            // if (outerPadding && !noOuterPadding && !isHeatmap && type !== 'clr' && !isPolar) {
            //     d3_outerPadding = scale.getOuterPadding(range, domain.length, padding, outerPadding);
            //     //bubble and datalabels
            //
            //     var range_correction = getRangeWithOuterPaddingCorrection(
            //         chartObj,
            //         type,
            //         index,
            //         domain,
            //         range,
            //         padding,
            //         outerPadding,
            //         d3_outerPadding
            //     );
            //     range = range_correction.range;
            //     outerPaddingPx = range_correction.outerPaddingPx;
            // }
            //
            // var scaleMode = type === 'clr' ? d3_scaleOrdinalOrg : d3_scaleOrdinal;
            // scaleObj = scaleMode().domain(domain);
            //
            // if (type === 'clr') {
            //     scaleObj.range(range);
            // } else {
            //     scaleObj
            //         .range(range)
            //         .paddingInner(padding)
            //         .paddingOuter(0);
            // }
            // // get and set all the range steps to ke xScale.ordinalRange
            // getOrdinalRange(scaleObj);
            // scaleObj.origRange = origRange;
            // scaleObj.outerPaddingPx = outerPaddingPx || 0;

            ////
            /*
                SCALE PARSER
            */
            //IMPORTANT: Padding is considered for single data. Refer /#/demoCharts/bullet1
            domain = type == 'clr' ? newScaleObj.domain() : domain;
            padding = domain.length === 1 ? 0 : padding;

            if (type == 'x' || type == 'y') {
                const ordinalAxisRangeParser = Registry.getComponent('axis.parser.ordinal');
                var scaleType = isPolarType(chartObj, type, index)
                        ? POLAR_SCALETYPE
                        : axis_data.ranges && axis_data.ranges.length
                        ? QUANTILE_SCALETYPE
                        : columnInfo.scaleType,
                    needCorrections = outerPadding && !noOuterPadding && !isHeatmap && type !== 'clr' && !isPolar,
                    rangeOperators = ordinalAxisRangeParser(type, index, chartObj, needCorrections, scaleType);

                // var orient = getAxisOrient(type, chartObj.axes.rotated);
                // rangeOperators.plotBound = {
                //     dimension: orient == BOTTOM_ALIGN || orient == TOP_ALIGN ? 'width' : 'height'
                // };
                // if (outerPadding && !noOuterPadding && !isHeatmap && type !== 'clr' && !isPolar) {
                //     if (
                //         (chartObj.datalabels.moduleExists && dataLabelEnabled(chartObj)) ||
                //         chartObj.dataObject.isBubbleEnabled ||
                //         DataProcessor.helpers.hasMarkerCorrection()
                //     ) {
                //         if (chartObj.dataObject.isBubbleEnabled) {
                //             rangeOperators.bubbleBound = {
                //                 mode: 'range'
                //             };
                //         } else if (DataProcessor.helpers.hasMarkerCorrection()) {
                //             rangeOperators.markerBound = {
                //                 mode: 'range'
                //             };
                //         } else if (labelRenderer) {
                //             rangeOperators.fontBound = {
                //                 mode: 'range'
                //             };
                //         }
                //     }
                // }

                var parser = {
                    dataType: datatype,
                    type: datatype,
                    scaleType,
                    field: columnInfo,
                    reverse: axis_data.reversed,
                    data: [],
                    domain: domain,
                    range: {
                        operators: rangeOperators,
                        evaluator: expressionEvalRangeForOrdinal
                    },
                    padding: padding,
                    outerPadding: outerPadding
                };
                if (chartObj.viewFlow) {
                    let _scale = chartObj.viewFlow.getScaleByColumnName(columnInfo, type);
                    if(_scale){
                        extendScaleParser(parser, _scale.conf, chartObj);
                    }
                    
                }
                // if (systemconf.viewParser) {
                //     if (systemconf.viewParser.scale && systemconf.viewParser.scale[columnInfo.name]) {
                //         extendScaleParser(parser, systemconf.viewParser.scale[columnInfo.name]);
                //     }
                // }
                var newScaleComp = new Scale(type, index, parser, chartObj);
                if (type !== 'clr') {
                    newScaleComp.d3_scale.paddingInner = parser.padding;
                    newScaleComp.d3_scale.paddingOuter = 0;
                }
                newScaleObj = newScaleComp.getScale();
            }
            // else {
            //     newScaleObj = newScaleObj;
            // }

            break;
        case 'ordinalNonLinear': {
            // init the scale with the leaf scale;

            var categories = dataObj.getCategories(type, index, applyZoom);
            domain = extent || dataObj.getDomain(categories, type);

            let rowList = hierarchical.pivotLayout.getRootNodeList('row'),
                columnList = hierarchical.pivotLayout.getRootNodeList('column');

            var origRange = [range[0], range[1]],
                rangeList;
            if (type == 'x' && columnList.getCount()) {
                rangeList = columnList.getRangesForBranch(index - columnList.getLeavesCount());
            } else if (type == 'y' && rowList.getCount()) {
                rangeList = rowList.getRangesForBranch(index - rowList.getLeavesCount());
            }

            scaleObj = nonLinearOrdinalScale()
                .domain(domain)
                .range(range)
                .rangeBands(rangeList);

            scaleObj.origRange = origRange;

            newScaleObj = scaleObj;
        }
    }

    if (chartObj.viewFlow) {
        let layout = getAxisLayout(orient),
            axisView = getAxisViewModel(chartObj, type, orient, index),
            viewParserScale = chartObj.viewFlow.getScaleByColumnName(columnInfo, type);
        if (axisView && axisView.scale && !chartObj.customVisual) {
            axisView.scale.component = newScaleObj;
        }else if(viewParserScale){
            viewParserScale.component = newScaleObj;
        }

    }

    // scaleObj.outerPadding = outerPadding;
    // scaleObj.padding = padding;
    // dataObj.hasNegativeValue[index] = newScaleObj.config.hasNegativeValue[index];
    newScaleObj.outerPadding = outerPadding;
    newScaleObj.padding = padding;

    return newScaleObj;
    /*
        SCALE PARSER
     */
    // return scaleObj;

    function getPadding() {
        if (isPolar || isHeatmap || type != 'x' || !chartObj.dataObject.isAxisCategory) {
            // Pivot with non axis category
            return 0;
        } else {
            let _default = dataObj.noofBarEnabledSeries
                ? 0.2
                : isFacet(chartObj) && Object.keys(dataObj.hierarchical.gridOptions).some(_ => hasColorColumnInNonAxis(chartObj, _))
                ? 0
                : 0.9;
            return validateNullVal(commonPlotOptions.padding, _default); //default value is 0.2 (Bar, Bullet, Gantt) or 0.9 (Others)
        }
    }

    function getOuterPadding() {
        if (
            isPolar ||
            hasColorScale ||
            isHeatmap ||
            noOuterPadding ||
            (!chartObj.dataObject.isAxisCategory && !chartObj.dataObject.isPolarAxisCategory && !chartObj.dataObject.isAxisCategory) ||
            isLogScale
        ) {
            // Pivot with non axis category
            return 0;
        } else if (type === 'y') {
            return validateNullVal(axis_data.outerPadding, chartObj.datalabels.hasFixedPosition ? 0.00000001 : 0.05);
        } else if (outerPadding != null) {
            return outerPadding;
        } else {
            return validateNullVal(commonPlotOptions.outerPadding, 0.02);
        }
    }
};

scale.setRowScale = function(chartObj, type, needCorrection) {
    var hierarchical = chartObj.dataObject.hierarchical,
        nodeList,
        nodeListCount,
        hasMultiCluster,
        yaxisObj = chartObj.axes.y,
        dataObject = chartObj.dataObject,
        plotarea = chartObj.plotarea,
        dataset = chartObj.dataset;

    if (hierarchical) {
        nodeList = hierarchical.pivotLayout.getRootNodeList('row');
        nodeListCount = nodeList.getCount();
        hasMultiCluster = nodeList.hasMultiCluster();
    }
    if (hasMultiCluster || nodeListCount) {
        hierarchical.pivotLayout.setRanges('row');

        var rowCount = nodeListCount || 0,
            clustersCount = nodeList.getClustersCount();
        if (rowCount || clustersCount) {
            var firstRowCategoryCount = rowCount ? nodeList.getDomainForBranch(0).length : 1,
                leavesCount = nodeList.getLeavesCount();
            /*
                Generate scale for branches
             */
            for (var j = 0; j < rowCount; j++) {
                if (!yaxisObj.branches[j]) {
                    yaxisObj.branches[j] = {};
                }
                let domain = nodeList.getDomainForBranch(j),
                    domainValues = dataObject.getDomain(domain, 'y', true);
                yaxisObj.branches[j].scale = scale.getScale(
                    chartObj,
                    dataObject,
                    plotarea,
                    yaxisObj.orient,
                    j + leavesCount,
                    dataset,
                    'y',
                    domainValues
                );
            }
            /*
            Generate scales for multi Y axis or clusters
             */
            var yscale = new Array(clustersCount);
            for (var i = 0; i < clustersCount; i++) {
                yscale[i] = scale.getScale(chartObj, dataObject, plotarea, yaxisObj.orient, i, dataset, 'y');
                nodeList.clusters[i].domain = yscale[i].domain();
            }
            var parentLeafRowIndex = 0,
                parentLeafRowScale = rowCount ? yaxisObj.branches[parentLeafRowIndex].scale : null;

            var tempRange = 0,
                co = 0;
            /*
                Generate Leaf scales
             */
            for (var k = 0; k < firstRowCategoryCount; k++) {
                for (var i = 0; i < clustersCount; i++) {
                    var axis_data = getAxisOptions(chartObj, 'y', i);
                    var ysc = yscale[i];
                    var rangeCopy = parentLeafRowIndex
                        ? [parentLeafRowScale(k) + parentLeafRowScale.discreteWidth(k), parentLeafRowScale(k)]
                        : [tempRange + nodeList.leafSize, tempRange];

                    tempRange = tempRange + nodeList.leafSize;
                    var correctionFactor = needCorrection ? 0.1 : 0,
                        diff = mathAbs(rangeCopy[1] - rangeCopy[0]),
                        correction = diff * correctionFactor;
                    rangeCopy = [rangeCopy[0] - correction, rangeCopy[1] + correction];

                    if ((chartObj.axes.rotated && !axis_data.reversed) || (!chartObj.axes.rotated && axis_data.reversed)) {
                        rangeCopy = rangeCopy.reverse();
                    }
                    if (!yaxisObj.leaves[co]) {
                        yaxisObj.leaves[co] = {};
                    }

                    yaxisObj.leaves[co].scale = ysc.copy().range(rangeCopy);
                    yaxisObj.leaves[co].scale.origRange = rangeCopy;
                    yaxisObj.leaves[co].parentBranchIdx = k;
                    co++;
                }
            }
            chartObj.axes.y = yaxisObj;
            return yaxisObj.leaves.concat(yaxisObj.branches);
        }
    }
};

scale.setColumnScale = function(chartObj, type) {
    var processor = chartObj.dataObject,
        hierarchical = processor.hierarchical,
        extent = chartObj.viewPortExtent,
        xaxes = chartObj.axes.x,
        dataObject = chartObj.dataObject,
        plotarea = chartObj.plotarea,
        dataset = chartObj.dataset,
        scaleFunc = scale.getScale,
        datatype = dataset.getDataType('x');
    /*
           set leafscale or rangescale for groupbyTime layout
            */
    xaxes.rangeScale = xaxes.scale.copy();
    xaxes.rangeScale.outerPaddingPx = xaxes.scale.outerPaddingPx;
    if (hierarchical.groupByDiscreteTime) {
        var categories = processor.getCategories('x'),
            domainValues = processor.getDomain(categories, 'x');
        xaxes.rangeScale.domain(domainValues);
    }

    let columnList = hierarchical.pivotLayout.getRootNodeList('column'),
        columnCount = columnList.getCount(),
        leafNodes =
            columnCount || datatype == TIME_DATATYPE || datatype == NUMERIC_DATATYPE
                ? columnList.getDomainForCluster(0)
                : dataObject.xaxisData.categories || columnList.getDomainForCluster(0),
        levels = columnCount || 0;
    /*
    set leafscale or rangescale for pivot or layer hierarchy layout
     */
    var domainValues = dataObject.getDomain(leafNodes, 'x', true);

    if (datatype == TIME_DATATYPE || datatype == NUMERIC_DATATYPE) {
        domainValues = columnCount ? d3_extent(domainValues) : xaxes.scale.domain(); // for time datatype, we didn't apply time format in pushToContinousCluster nodes, hence using existing domain
    }
    chartObj.axes.x.leafScaleGrid = hierarchical.categoryTree
        ? scale.getScale(chartObj, dataObject, plotarea, xaxes.orient, null, dataset, 'x', domainValues)
        : chartObj.axes.x.scale;

    if (columnCount) {
        var nodeList = hierarchical.pivotLayout.getRootNodeList('column');
        if (chartObj.systemConf.facets && chartObj.systemConf.facets.config && chartObj.systemConf.facets.config.columns.leafSize) {
            chartObj.dataObject.hierarchical.getDynamicColumnSize(chartObj.axes.x.leafScaleGrid);
            var dynamicDomain = nodeList.getDynamicDomainForCluster(0);
            var xsc1 = scale.getScale(chartObj, dataObject, plotarea, xaxes.orient, null, dataset, 'x', d3_range(dynamicDomain.length));
            var leafRanges = hierarchical.pivotLayout.setRangesForLeaf('column', xsc1);

            var newscaleObj = nonLinearOrdinalScale()
                .domain(domainValues)
                .range(xsc1.origRange)
                .rangeBands(leafRanges)
                .rangeBand(xsc1.bandwidth());

            newscaleObj.origRange = xsc1.origRange;
            newscaleObj.outerPaddingPx = xaxes.scale.outerPaddingPx;
            xaxes.rangeScale = newscaleObj;
            xaxes.rangeScale.outerPaddingPx = newscaleObj.outerPaddingPx;
            xaxes.scale = newscaleObj;
            xaxes.scale.outerPaddingPx = newscaleObj.outerPaddingPx;
            chartObj.axes.x.leafScaleGrid = newscaleObj;
            chartObj.axes.x.leafScaleGrid.outerPaddingPx = newscaleObj.outerPaddingPx;
            // chartObj.axes.x.dynamicScale = newscaleObj;
        }
        /*
    Find ranges for column branches
     */
        let rangeScaleUtils = hierarchical.setScaleForRange(chartObj.axes.x.rangeScale);
        hierarchical.pivotLayout.setRanges('column', rangeScaleUtils);
        /*
        Generate scale for column branches
         */
        for (var i = 0; i < levels; i++) {
            if (!xaxes.levels[i]) {
                xaxes.levels[i] = {};
            }
            xaxes.levels[i].scale = scaleFunc(
                chartObj,
                processor,
                plotarea,
                xaxes.orient,
                i + columnList.getLeavesCount(),
                dataset,
                'x',
                extent,
                null
            );
        }
    }
    /*
       Find leaf scales
        */
    var hasSplitLeavesColumns = chartObj.systemConf.chart.axes.xaxis.splitLeaves,
        reversed = chartObj.systemConf.chart.axes.xaxis.reversed,
        firstColumnCategoryCount = columnList.getLeavesCount();

    if (hasSplitLeavesColumns && columnCount) {
        var uniformLeafScale = chartObj.systemConf.chart.axes.xaxis.uniformLeafScale,
            uniqueDomainList = datatype == ORDINAL_DATATYPE ? columnList.getUniqueDomainForCluster(0) : null;
        var leafScale = scale.getScale(chartObj, dataObject, plotarea, xaxes.orient, null, dataset, 'x', uniqueDomainList);
        var parentLeafColumnIndex = 0,
            parentLeafColumnScale = columnCount ? xaxes.levels[parentLeafColumnIndex].scale : null;
        var isHorizontal = xaxes.orient == 'top' || xaxes.orient == 'bottom',
            plotRangeSign = isHorizontal ? 1 : -1,
            tempRange = isHorizontal ? 0 : plotarea.height,
            co = 0;

        //calculate outerpaddingpx for leafscale
        var leafscale_outerpaddingPX;

        for (var k = 0; k < firstColumnCategoryCount; k++) {
            var domain = datatype == ORDINAL_DATATYPE ? columnList.getDomainForClusterSet(0, k) : null,
                leafSize = uniformLeafScale ? columnList.leafSize : columnList.nonUniformLeafSize[k],
                rangeCopy = [tempRange, tempRange + plotRangeSign * leafSize];
            tempRange = tempRange + plotRangeSign * leafSize;
            var correctionFactor = 0.1,
                diff = mathAbs(rangeCopy[1] - rangeCopy[0]),
                correction = diff * correctionFactor,
                startRangeSign = isHorizontal ? 1 : -1,
                endRangeSign = isHorizontal ? -1 : 1;
            rangeCopy = [rangeCopy[0] + startRangeSign * correction, rangeCopy[1] - startRangeSign * correction];
            if (reversed) {
                rangeCopy = rangeCopy.reverse();
            }
            if (!xaxes.leaves[co]) {
                xaxes.leaves[co] = {};
            }

            var leafScaleCopy = uniformLeafScale
                ? leafScale
                : scale.getScale(chartObj, dataObject, plotarea, xaxes.orient, null, dataset, 'x', domain);

            xaxes.leaves[co].scale = leafScaleCopy.copy().range(rangeCopy);
            xaxes.leaves[co].scale.origRange = rangeCopy;
            xaxes.leaves[co].parentBranchIdx = k;

            if (!leafscale_outerpaddingPX) {
                var leafscale_domain = leafScale.domain(),
                    padding = leafScale.padding,
                    outerPadding = leafScale.outerPadding;
                var d3_outerPadding = scale.getOuterPadding(rangeCopy, leafscale_domain.length, padding, outerPadding);
                var range_correction = getRangeWithOuterPaddingCorrection(
                    chartObj,
                    'x',
                    0,
                    leafscale_domain,
                    rangeCopy,
                    padding,
                    outerPadding,
                    d3_outerPadding
                );
                xaxes.leaves[co].scale.outerPaddingPx = range_correction.outerPaddingPx || 0;
            }
            co++;
        }
        return xaxes.leaves.concat(xaxes.levels);
    }

    return [
        {
            scale: chartObj.axes.x.leafScaleGrid
        }
    ].concat(xaxes.levels);
};
scale.getEventScale = function(chartObj, dataObj, plotarea, orient, index, dataset, type) {
    var datatype = dataset.getDataType(type, index);
    var scaleObj;
    switch (datatype) {
        default:
            scaleObj = scale.getScale(chartObj, dataObj, plotarea, orient, index, dataset, type); //No I18N ;
    }
    return scaleObj;
};
scale.constructor = function(chartObj, orient, scaleFunc, eventscaleFunc) {
    var scaleObj;
    var defaultscaleFunc = scale.getScale,
        defaulteventscaleFunc = scale.getEventScale,
        scaleFunc = pick(scaleFunc, defaultscaleFunc),
        eventscaleFunc = pick(eventscaleFunc, defaulteventscaleFunc);

    switch (orient) {
        case 'left':
            scaleObj = chartObj.axes.rotated ? getxscale() : getyscale();
            break;
        case 'bottom':
            scaleObj = chartObj.axes.rotated ? getyscale() : getxscale();
            break;
    }
    return scaleObj;

    function getxscale() {
        var processor = chartObj.dataObject,
            hierarchical = processor.hierarchical,
            extent = chartObj.viewPortExtent,
            xaxes = chartObj.axes.x,
            dataObject = chartObj.dataObject,
            plotarea = chartObj.plotarea,
            dataset = chartObj.dataset;

        chartObj.axes.x.scale = scaleFunc(chartObj, dataObject, plotarea, xaxes.orient, 0, dataset, 'x', extent);
        chartObj.axes.x.eventscale =
            eventscaleFunc == null
                ? chartObj.axes.x.scale
                : eventscaleFunc(chartObj, dataObject, plotarea, xaxes.orient, null, dataset, 'x'); //No I18N
        // if(extent){
        //     chartObj.viewPortExtent = chartObj.axes.x.scale.domain();
        //     if(chartObj.scroll.scrollDomain){
        //         chartObj.scroll.scrollDomain = chartObj.viewPortExtent;
        //     }
        // }
        if (defined(hierarchical)) {
            /*
            set leafscale or rangescale for groupbyTime layout
             */
            return scale.setColumnScale(chartObj, 'x');
        }

        return [
            {
                scale: chartObj.axes.x.scale
            }
        ].concat(xaxes.levels);
    }

    function getyscale() {
        var processor = chartObj.dataObject,
            hierarchical = processor.hierarchical;
        var splityaxes = hierarchical && hierarchical.categoryTree ? 1 : chartObj.axes.y.count;
        var yaxes = chartObj.axes.y;
        for (var i = 0; i < splityaxes; i++) {
            if (!yaxes.groups[i]) {
                yaxes.groups[i] = {};
            }
            yaxes.groups[i].scale = scaleFunc(
                chartObj,
                chartObj.dataObject,
                chartObj.plotarea,
                chartObj.axes.y.orient,
                i,
                chartObj.dataset,
                'y'
            );
            yaxes.groups[i].eventscale =
                eventscaleFunc == null
                    ? yaxes.groups[i].scale
                    : eventscaleFunc(chartObj, chartObj.dataObject, chartObj.plotarea, chartObj.axes.y.orient, i, chartObj.dataset, 'y');
        }

        if (hierarchical && hierarchical.categoryTree) {
            let rowList = hierarchical.pivotLayout.getRootNodeList('row'),
                rowCount = rowList.getCount(),
                hasMultiCluster = rowList.hasMultiCluster();

            if (rowCount || hasMultiCluster) {
                if (chartObj.axes.rotated) {
                    scale.setRowScale(chartObj, 'row', true);
                }
                if (rowCount) {
                    return chartObj.axes.y.leaves.concat(chartObj.axes.y.branches);
                }
                return chartObj.axes.y.leaves;
            }
            return chartObj.axes.y.groups;
        }
        return chartObj.axes.y.groups;
    }
};
function getSurroundingPoints(args, sizeCallback) {
    var { datatype, point, chartObj, axisrotated, scaleObj, type, xidx } = args,
        quadTree,
        chartCache = chartObj.cache,
        dataInfo = chartCache.brushDataDetails,
        cacheQuadTree = dataInfo && dataInfo[datatype].quadTree,
        accessor = chartObj.eventHandler && chartObj.eventHandler.dataAccesor([point, 0]),
        paneid = accessor && `pane${accessor.paneId()}`,
        flattenPoints = chartObj && chartObj.dataObject && chartObj.dataObject.flattenPoints,
        flattendata = flattenPoints && flattenPoints.get(paneid),
        isRedraw = chartObj && chartObj.updating,
        nearestDataPoints;

    if (dataInfo) {
        chartCache.brushDataDetails[datatype].scale = scaleObj;
        chartCache.brushDataDetails[datatype].brushType = type;
        chartCache.brushDataDetails[datatype].activeBrush = chartObj.eventHandler.activeBrushComponent;
    }

    if ((dataInfo && !cacheQuadTree && flattendata) || (dataInfo && isRedraw && flattendata)) {
        var QuadTree = Registry.getComponent('searchMode.quadtree'),
            keys = flattendata.keys(),
            data = [];

        //check this case
        chartObj.systemConf.chart.plot.events.enabled = undefined;
        keys.map(key => {
            let d = flattendata.get(key),
                a = objectToArray(
                    d,
                    d0 => {
                        return !d0.disabled();
                    },
                    accessor.spliceIndices()
                );
            data = data.concat(a[0] instanceof Array ? d3_merge(a) : a);
        });
        quadTree = new QuadTree(data, true, Infinity, [chartObj.plotarea.width, chartObj.plotarea.height], null);
        quadTree.tree
            .x(function(d) {
                var scaleObj = chartCache.brushDataDetails[datatype].scale,
                    yidx = defined(d.yaxisOrder) ? chartObj.dataset.yidx(d.yaxisOrder) : d.yColumnIndex,
                    type = chartCache.brushDataDetails[datatype].brushType,
                    _val = type === 'y' ? d[yidx] : d[xidx];
                if ((axisrotated && type === 'x') || (!axisrotated && type === 'y')) {
                    return 0;
                }
                return scaleObj(_val) + (scaleObj.bandwidth != null ? scaleObj.bandwidth(_val) / 2 : 0);
            })
            .y(function(d) {
                var scaleObj = chartCache.brushDataDetails[datatype].scale,
                    yidx = defined(d.yaxisOrder) ? chartObj.dataset.yidx(d.yaxisOrder) : d.yColumnIndex,
                    type = chartCache.brushDataDetails[datatype].brushType,
                    _val = type === 'y' ? d[yidx] : d[xidx];
                if (!((axisrotated && type === 'x') || (!axisrotated && type === 'y'))) {
                    return 0;
                }
                return scaleObj(_val) + (scaleObj.bandwidth != null ? scaleObj.bandwidth(_val) / 2 : 0);
            })
            .size(sizeCallback);
        chartCache.brushDataDetails[datatype].quadTree = quadTree;
        chartCache.brushDataDetails[datatype].data_sizePos = d3_map();
        cacheQuadTree = quadTree;
    }
    // quadTree = cacheQuadTree;
    nearestDataPoints =
        cacheQuadTree && cacheQuadTree.find((axisrotated && type === 'x') || (!axisrotated && type === 'y') ? [0, point] : [point, 0]);
    return nearestDataPoints || [];
}
function getClosestPoint(args, nearestDataPoints, selection_Start, selection_End) {
    var { datatype, point, chartObj, ordinalRange, scaleObj, maxBandwidth, rangeBand, type, xidx } = args,
        yidx,
        nearestData_pos,
        nearestData_size,
        xVal,
        sizePosMap = chartObj.cache.brushDataDetails[datatype].data_sizePos,
        _minDiff,
        activeBrush = chartObj.eventHandler.activeBrushComponent;
    
    for (var i = 0; i < nearestDataPoints.length; i++) {
        var _p = nearestDataPoints[i];
        yidx = chartObj.dataset.yidx(_p.yaxisOrder);
        var _val = type === 'y' ? _p[yidx] : _p[xidx],
            _p_info = sizePosMap.get(_val);

        if (_p_info && !(activeBrush === "plot")) {
            //we avoid the saved values in brush cases, since the position will be varying for each bar/marker in case of plot brush
            //collect it from cache if saved already
            if (selection_Start) {
                if (_p_info.pos < nearestData_pos) {
                    nearestData_pos = _p_info.pos;
                    nearestData_size = _p_info.size;
                    xVal = _val;
                }
            } else {
                if (_p_info.pos > nearestData_pos) {
                    nearestData_pos = _p_info.pos;
                    nearestData_size = _p_info.size;
                    xVal = _val;
                }
            }
        }
        var getDiff = (a, b) => {
            var max = Math.max(a, b),
                min = Math.min(a, b);
            return max - min;
        };

        var _p_size = _p._size || (rangeBand > maxBandwidth ? maxBandwidth : rangeBand),
            _p_pos = scaleObj(_val) + (rangeBand ? rangeBand / 2 : 0),
            _d = getDiff(point, _p_pos);
        _minDiff = isUndefined(_minDiff) ? _d : _minDiff;
        if (_d < _minDiff || isUndefined(nearestData_pos)) {
            _minDiff = _d;
            nearestData_pos = _p_pos;
            xVal = _val;
        }
        nearestData_size = isUndefined(nearestData_size) || _p_size >= nearestData_size ? _p_size : nearestData_size; //to collect largest marker size
    }

    if (!sizePosMap.get(xVal) && !(activeBrush === "plot")) {
        sizePosMap.set(xVal, { pos: nearestData_pos, size: nearestData_size });
    }

    if (defined(nearestData_pos) && defined(nearestData_size)) {
        nearestData_pos = nearestData_pos - nearestData_size / 2;
        if (point >= nearestData_pos && point <= nearestData_pos + nearestData_size) {
            if (datatype === 'ordinal') {
                var itemIndex_obj;
                if (ordinalRange) {
                    itemIndex_obj = scaleObj.domain().reduce((t, d, i) => ((t[d] = i), t), {});
                }
                return itemIndex_obj[xVal];
            } else {
                return selection_Start ? nearestData_pos : selection_End ? nearestData_pos + nearestData_size : point;
            }
        }
    }
}
scale.invert = function(datatype, mousePos, scaleObj, axisrotated, type, axisReversed, useWithinBound, brushSelection, chartObj) {
    var actualPos;
    switch (datatype) {
        case 'numeric':
        case 'time':
            var point = type == 'y' ? (axisrotated ? mousePos.x : mousePos.y) : axisrotated ? mousePos.y : mousePos.x;
            var _args = {
                    chartObj: chartObj,
                    axisrotated: chartObj.axes.axisrotated,
                    datatype: datatype,
                    point: point,
                    scaleObj: scaleObj,
                    type: type,
                    xidx: chartObj.systemConf.metadata.columns[chartObj.systemConf.metadata.axes.x[0]].dataindex
                },
                chartCache = chartObj.cache,
                cacheDataInfo = chartCache.brushDataDetails && chartCache.brushDataDetails[datatype];
            if (!cacheDataInfo) {
                chartCache.brushDataDetails = {};
                chartCache.brushDataDetails[datatype] = {};
            }
            var sizeCallback = function(d) {
                    let data = defined(d._size) ? d : d._data(),
                        chart =
                        chartCache.brushDataDetails[datatype].activeBrush === 'scroll'
                                ? chartObj.overview && chartObj.overview.destroying
                                    ? chartObj
                                    : chartObj.overview || chartObj
                                : chartObj;
                    if (chart.scales.length) {
                        return getSizeFromPoint(d, chartObj.dataset, chart);
                    }
                    return data._size;
                },
                selection_Start = brushSelection && brushSelection.selection && brushSelection.selection[0] === point,
                selection_End = brushSelection && brushSelection.selection && brushSelection.selection[1] === point,
                oldBrush = chartObj.eventHandler.oldBrushComponent,
                activeBrush = chartObj.eventHandler.activeBrushComponent,
                findNearestPoint = (oldBrush === 'plot' && activeBrush === 'scroll') || (oldBrush === null && activeBrush === null),
                isOverview = chartObj.scroll && chartObj.scroll.overviewEnabled;

            if (!findNearestPoint && (activeBrush === 'plot' || isOverview)) {
                var nearestDataPoints = getSurroundingPoints(_args, sizeCallback);
                if (nearestDataPoints.length) {
                    var _ = getClosestPoint(_args, nearestDataPoints, selection_Start, selection_End);
                    point = defined(_) ? _ : point;
                }
            }

            actualPos = scaleObj.invert(point);
            break;
        case 'ordinal':
            var ordinalRange = getOrdinalRange(scaleObj),
                rangeBand = scaleObj.bandwidth(),
                maxBandwidth = scaleObj.maxBandWidth1,
                actualWidth = maxBandwidth ? mathMin(rangeBand, maxBandwidth) : rangeBand,
                diffRange = ordinalRange.length > 1 ? mathAbs(ordinalRange[1] - ordinalRange[0]) : 0,
                actualPadding = mathAbs(diffRange - actualWidth),
                padding_maxBandwidth = maxBandwidth && maxBandwidth < rangeBand ? (rangeBand - actualWidth) / 2 : 0,
                sign = axisrotated && !axisReversed ? 1 : -1,
                correction = axisrotated && !axisReversed ? actualWidth : 0,
                validator = (axisrotated && !axisReversed) ? '<' : '>',
                outspace = 0,
                actualPos,
                chartCache = chartObj.cache,
                cacheDataInfo = chartCache.brushDataDetails && chartCache.brushDataDetails.ordinal;
                // isScrollChartWithoutOverview = chartObj.eventHandler.activeBrushComponent === "scroll" && !chartObj.overview;

            // if(isScrollChartWithoutOverview){
            //     if (type === 'x') {
            //         maxBandwidth = chartObj.axes.x.scale.maxBandWidth;
            //     } else {
            //         maxBandwidth = chartObj.axes.y.groups[index].scale.maxBandWidth;
            //     }
            // }
            if (maxBandwidth && rangeBand > maxBandwidth) {
                outspace = (rangeBand - maxBandwidth) / 2;
            }

            ordinalRange = ordinalRange.map(function(d) {
                if (brushSelection) {
                    return d + outspace;
                    //why this part?
                    //outspace - when maxWidth applied the starting of the bar will not be proper, so we add up that offset
                    //why (sign * actualPadding) / 2 not used?
                    //if its used, we can't get the start of the bar exactly, we don't need an extra space on start and end of the bar.
                }
                return d + correction + (sign * actualPadding) / 2;
                //why no outspace here?
                //here (sign * actualPadding) / 2 is the extra space on start and end of the bar
                // if outspace used, the space got on start and end of the bar would be high.
            });
            if (!cacheDataInfo) {
                chartCache.brushDataDetails = {};
                chartCache.brushDataDetails[datatype] = {};
            }
            var point = type == 'y' ? (axisrotated ? mousePos.x : mousePos.y) : axisrotated ? mousePos.y : mousePos.x,
                _args = {
                    chartObj: chartObj,
                    axisrotated: chartObj.axes.rotated,
                    axisReversed: axisReversed,
                    datatype: datatype,
                    point: point,
                    ordinalRange: ordinalRange,
                    maxBandwidth: maxBandwidth,
                    rangeBand: rangeBand,
                    scaleObj: scaleObj,
                    type: type,
                    xidx: chartObj.dataSplicer.getColumns()[chartObj.dataSplicer.getEncode()["x"][0]].dataindex
                },
                sizeCallback = function(d) {
                    if (d._width) {
                        var scaleObj = chartCache.brushDataDetails[datatype].scale;
                        return scaleObj.bandwidth() > scaleObj.maxBandWidth1 ? scaleObj.maxBandWidth1 : scaleObj.bandwidth();
                    }
                    let data = defined(d._size) ? d : d._data(),
                        chart =
                        chartCache.brushDataDetails[datatype].activeBrush === 'scroll'
                                ? chartObj.overview && chartObj.overview.destroying
                                    ? chartObj
                                    : chartObj.overview || chartObj
                                : chartObj;
                    if (chart.scales.length) {
                        return getSizeFromPoint(d, chartObj.dataset, chart);
                    }
                    return data._size;
                },
                selection_Start = brushSelection && brushSelection.selection && brushSelection.selection[0] === point,
                selection_End = brushSelection && brushSelection.selection && brushSelection.selection[1] === point,
                oldBrush = chartObj.eventHandler.oldBrushComponent,
                activeBrush = chartObj.eventHandler.activeBrushComponent,
                findNearestPoint = (oldBrush === 'plot' && activeBrush === 'scroll') || (oldBrush === null && activeBrush === null);
            if (!findNearestPoint && ((oldBrush && oldBrush.length) || (activeBrush && activeBrush.length))) {
                var nearestDataPoints = getSurroundingPoints(_args, sizeCallback);
                if (nearestDataPoints.length) {
                    actualPos = getClosestPoint(_args, nearestDataPoints, selection_Start, selection_End);
                }
            }

            var quadTree = chartObj && chartObj.cache && chartObj.cache.brushDataDetails[datatype].quadTree;
            if (brushSelection && brushSelection.quadTreePosition && quadTree) {
                brushSelection.quadTreePosition[selection_Start ? 0 : 1] = actualPos;
            }
            if (defined(actualPos)) {
                break;
            }
            if ((oldBrush && oldBrush.length) || (activeBrush && activeBrush.length)) {
                //this check is needed since scale.invert iss used for FEATURES_BAR_25 triggeroffset
                ordinalRange = getOrdinalRange(scaleObj).map(function(d) {
                    //getting the middle position of each rect/marker
                    d = d + scaleObj.bandwidth() / 2;
                    return d;
                });
            }
            actualPos = getBisectedValue(mousePos, ordinalRange, type, axisrotated, axisReversed, validator);
            var x = scaleObj.ordinalRange[actualPos] + padding_maxBandwidth;
            if (useWithinBound && !isWithinRange(point, [x, x + actualWidth])) {
                actualPos = null;
            }
            if (brushSelection) {
                actualPos = adjustActualPos(actualPos, ordinalRange, selection_Start, selection_End, _args);
            }
            break;
    }
    return actualPos;
};
function adjustActualPos(actPos, ordinalRange, selection_Start, selection_End, _args) {
    var { axisReversed, axisrotated, type, point } = _args;
    var curr_Pos = ordinalRange[actPos],
        val = point,
        _ = (actPos, pos) => {
            if (defined(ordinalRange[pos])) {
                return pos;
            }
            return actPos;
        };
    if (selection_Start) {
        if (
            (type === 'x' && ((axisrotated && !axisReversed) || (!axisrotated && axisReversed))) ||
            (type === 'y' && ((!axisrotated && !axisReversed) || (axisrotated && axisReversed)))
        ) {
            if (val > curr_Pos) {
                actPos = _(actPos, --actPos);
            }
        } else {
            if (val > curr_Pos) {
                actPos = _(actPos, ++actPos);
            }
        }
    } else if (selection_End) {
        if (
            (type === 'x' && ((axisrotated && !axisReversed) || (!axisrotated && axisReversed))) ||
            (type === 'y' && ((!axisrotated && !axisReversed) || (axisrotated && axisReversed)))
        ) {
            if (val < curr_Pos) {
                actPos = _(actPos, ++actPos);
            }
        } else {
            if (val < curr_Pos) {
                actPos = _(actPos, --actPos);
            }
        }
    }
    return actPos;
}

scale.ordinalSubInvert = function(mousePos, itemPos, scaleObj, primaryScale, axisrotated, type, axisReversed) {
    var rangeBand = scaleObj.bandwidth(),
        range = scaleObj.range(),
        primaryRange = primaryScale.ordinalRange,
        domain = scaleObj.domain(),
        range_diff = range[1] - range[0],
        padding = range_diff - domain.length * rangeBand,
        ordinalRange = domain.map(function(d, i) {
            return primaryRange[itemPos] + i * (rangeBand + padding);
        }),
        insertionPos = getBisectedValue(mousePos, ordinalRange, type, axisrotated, axisReversed);
    return insertionPos;
};
export function getBisectedValue(mousePos, arr, type, axisrotated, axisReversed, validator) {
    var bisectVal = type == 'y' ? (axisrotated ? mousePos.x : mousePos.y) : axisrotated ? mousePos.y : mousePos.x,
        orient = getAxisOrient(type, axisrotated),
        defaultReverse = orient == LEFT_ALIGN,
        rangeReversed = axisReversed ? !defaultReverse : defaultReverse,
        comparator = rangeReversed ? d3_descending : d3_ascending,
        endPos = arr.length - 1;
    return bisectArray(bisectVal, arr, comparator, endPos, validator, rangeReversed);
}
scale.barBandwidth = function(chartObj) {
    var systemconf = chartObj.systemConf,
        dataObj = chartObj.dataObject;
    var temp = getFirstAxisChartType(chartObj),
        allChartTypes = chartObj.dataset.seriesTypes,
        barType =
            dataObj.barChartType ||
            (dataObj.noofBarEnabledSeries
                ? allChartTypes.has(charttype.stkdbar)
                    ? charttype.stkdbar
                    : allChartTypes.has(charttype.bar)
                    ? charttype.bar
                    : allChartTypes.has(charttype.bullet)
                    ? charttype.bullet
                    : allChartTypes.has(charttype.waterfall)
                    ? charttype.waterfall
                    : temp
                : temp);
    dataObj.barChartType = barType;
    // temp = barType || temp;
    let axisView = getAxisViewModel(chartObj, 'x', chartObj.axes.x.orient, 0),
        subBandCount;
    if (axisView && axisView.scale) {
        subBandCount = axisView.scale.conf.subBand;
    }
    var m;
    if (dataObj.isstacked(barType)) {
        //#ZC1346
        var stackTransform = dataObj.getEncode('stack'),
            check = stackData => isBarEnabled.call(dataObj, stackData.seriesData[0]);

        m = dataObj.categoryHierarchy ? 1 : objectValues(stackTransform).filter(check).length || 1; // length = 0 for continuous chart with ordinal scale
    } else if (dataObj.hierarchical && dataObj.hierarchical.pivotLayout.getRootNodeList('row').hasMultiCluster()) {
        m = 1;
    } else {
        m = pick(subBandCount, getBandCount(chartObj));
    }

    var cType = charttypenames.get(temp);
    var isBullet = temp == charttype.bullet,
        isHeatmap = isHeatMapChartType(temp);

    var plotBorderOptions = systemconf.chart.plot.border;
    var borderSize = plotBorderOptions.show === true ? plotBorderOptions.size : 0;
    var padding =
        cType == 'gantt' || cType == 'bullet' || cType == 'bar' || cType == 'waterfall' || cType == 'boxplot' || cType == 'barrange'
            ? systemconf.chart.plot.plotoptions[cType].padding || 0.2
            : 0.2;
    var interPadding =
        m == 1
            ? 0
            : cType == 'bar' || cType == 'gantt' || cType == 'waterfall' || cType == 'boxplot' || cType == 'barrange'
            ? systemconf.chart.plot.plotoptions[cType].interPadding || 0
            : 0;

    var outerPadding = systemconf.chart.plot.plotoptions[cType].outerPadding;
    outerPadding = outerPadding != null && outerPadding !== '' ? outerPadding : 0.02;
    var n = chartObj.axes.x.scale.domain().length; //chart.dataObject.getCategories("x").length;
    var range = !chartObj.axes.rotated
            ? [borderSize, getPlotSize(chartObj, 'width', chartObj.plotarea) - borderSize]
            : [getPlotSize(chartObj, 'height', chartObj.plotarea) - borderSize, borderSize],
        barscale1,
        barscale2;

    outerPadding = scale.getOuterPadding(range, n, padding, outerPadding);

    chartObj.barscale1 = barscale1 = chartObj.axes.x.leaves.length ? chartObj.axes.x.leaves[0].scale : chartObj.axes.x.scale;
    chartObj.barscale2 = barscale2 = d3_scaleOrdinal()
        .domain(d3_range(m))
        .range([0, barscale1.bandwidth()])
        .paddingInner(interPadding);

    //setting bandwidth if the actual bandwidth is greater than the maximum bar bandwidth
    // barscale2.rangeBands([0, barscale1.bandwidth()], interPadding);
    var maxBandWidth_cp = invokeFunction(systemconf.chart.plot.plotoptions[cType].maxBandWidth, barscale1, chartObj);
    maxBandWidth_cp = maxBandWidth_cp < 0 ? null : maxBandWidth_cp;
    var maxBandWidth = dataObj.noofBarEnabledSeries && !isBullet ? maxBandWidth_cp || barscale1.bandwidth() : 200;

    var maxBandWidth1 = invokeFunction(
        systemconf.chart.plot.plotoptions[charttypenames.get(barType || temp)].maxBandWidth,
        barscale1,
        chartObj
    );
    maxBandWidth1 = maxBandWidth1 < 0 ? null : maxBandWidth1;

    barscale1.maxBandWidth = m * maxBandWidth_cp;
    barscale1.maxBandWidth1 = m * maxBandWidth1;
    if (!isBullet && barscale2.bandwidth() > maxBandWidth) {
        var outspace = (barscale1.bandwidth() - m * maxBandWidth) / 2;
        barscale2.range([outspace, barscale1.bandwidth() - outspace]).paddingInner(interPadding);
    }
    scale.outspace = barscale2.range()[0];
    return [barscale1, barscale2];
};

scale.getOuterPadding = function(range, domainLength, padding, outerPadding) {
    var rangeDiff = mathAbs(range[1] - range[0]);
    var pad = outerPadding * rangeDiff;
    var step = (rangeDiff - 2 * pad) / (domainLength - padding + 2 * 0);
    return pad / step;
};

scale.getOrdinalOuterPadding = function(range, domainLength, padding, outerPadding) {
    var rangeDiff = mathAbs(range[1] - range[0]);
    var pad = outerPadding * rangeDiff;
    return pad / 2;
};

/**
 *
 * @param scale linear scale
 * @param points tick values
 * @returns {{outerPaddingInPixel, rangeBand: number}}
 */

scale.getContinuousRangeBand = function(scaleObj, points, index) {
    var ranges = scaleObj.range(),
        firstPointSpan = scaleObj(points[0]),
        currentPointSpan = index && points[index] ? scaleObj(points[index]) : firstPointSpan,
        nextPointSpan = index && points[index + 1] ? scaleObj(points[index + 1]) : scaleObj(points[1] || points[0]),
        rangeBand = points.length == 1 ? (firstPointSpan - (firstPointSpan - ranges[0]) / 2) * 2 : nextPointSpan - currentPointSpan,
        // for axes rotated or axes reversed.
        correction = rangeBand < 0 ? -1 : 1;

    return {
        outerPaddingInPixel: correction * mathMin(mathAbs(firstPointSpan - ranges[0]), mathAbs(rangeBand / 2)),
        rangeBand: rangeBand / 2
    };
};
scale.getDiscreteRangeBand = function(scaleObj) {
    var rangeBand = scaleObj.bandwidth(),
        // rangeExtent = scaleObj.range(),
        outerPadding = scaleObj.outerPaddingPx + rangeBand / 2, //(scaleObj.range()[0] - rangeExtent[0])+(rangeBand/2),
        ordinalRange = getOrdinalRange(scaleObj),
        diffBand = scaleObj.domain().length > 1 ? ordinalRange[1] - ordinalRange[0] - rangeBand : 0,
        value_1 = rangeBand / 2 + diffBand / 2;
    return {
        outerPaddingInPixel: mathMin(outerPadding, value_1),
        rangeBand: value_1
    };
};
/*
TODO Need to check both these function differs in value for hierarchy and axis
 */
scale._getDiscreteRangeBand = function(scaleObj, index) {
    var rangeBand = scaleObj.bandwidth(),
        outerPadding = scaleObj.outerPaddingPx,
        ordinalRange = getOrdinalRange(scaleObj),
        hasOrdinalRange = isArray(ordinalRange) && ordinalRange.length > 1,
        startIdx = pick(index, 0),
        endIdx;
    if (hasOrdinalRange && startIdx == ordinalRange.length - 1) {
        //sort the indices in ascending order startIdx to endIdx
        endIdx = startIdx;
        startIdx = startIdx - 1;
    } else {
        endIdx = startIdx + 1;
    }
    // endIdx = hasOrdinalRange && startIdx == ordinalRange.length - 1 ? startIdx - 1 : startIdx + 1,
    var value_1 = hasOrdinalRange ? (ordinalRange[endIdx] - (ordinalRange[startIdx] + rangeBand)) / 2 : 0; // #ZC1156
    return {
        outerPaddingInPixel: mathMin(outerPadding, value_1),
        rangeBand: value_1
    };
};
scale.getRangeBand = function(scaleObj, points, index) {
    var rb;
    if (scaleObj.bandwidth) {
        rb = scale.getDiscreteRangeBand(scaleObj);
    } else {
        rb = scale.getContinuousRangeBand(scaleObj, points, index);
    }
    return rb;
};

export function getCategoryStep(rangeband, padding) {
    return rangeband / (1 - padding);
}

export function getOuterPaddingInPixel(plotSize, count, padding, userOuterPadding, step) {
    var outerPadding = scale.getOuterPadding(plotSize, count, padding, userOuterPadding),
        outerPaddingPx = 2 * (mathMin(plotSize[0], plotSize[1]) + step * outerPadding);
    return outerPaddingPx; 
}

export function getRangeWithOuterPaddingCorrection(chartObj, type, index, domain, range, padding, userOuterPadding, userOuterPadding_d3) {
    var outerPaddingPx = 0,
        outerPaddingPx_others = 0,
        outerPaddingPx_user = 0;
    var min2Max = range[0] < range[1];
    var minPos = min2Max ? 0 : 1,
        maxPos = min2Max ? 1 : 0;

    // For bubble, datalabels, marker correction
    outerPaddingPx_others = chartObj.dataObject.extendOuterpadding(
        domain,
        range,
        chartObj.systemConf,
        padding,
        userOuterPadding_d3,
        type,
        index
    );
    outerPaddingPx += outerPaddingPx_others;

    //For user given outerpadding
    var diff = mathAbs(range[1] - range[0]);
    outerPaddingPx_user = diff * userOuterPadding;
    outerPaddingPx += outerPaddingPx_user;

    //Outerpadding corrections in range
    range[minPos] = range[minPos] + outerPaddingPx;
    range[maxPos] = range[maxPos] - outerPaddingPx;

    return { range: range, outerPaddingPx: outerPaddingPx };
}

/**
 * it's a function to receive all ranges of a ordinal scale equivalent to d3_ordinal_scale.range() in ~d3.v3
 * @param scaleObj d3 ordinal scale
 * @returns {[Number]}
 */
export function getOrdinalRange(scaleObj) {
    return scaleObj.ordinalRange || (scaleObj.ordinalRange = scaleObj.domain().map(d => scaleObj(d)));
}

export function getActualRange(scaleObj) {
    return scaleObj.actualRange || (scaleObj.actualRange = scaleObj.originalDomain && scaleObj.originalDomain.map(scaleObj));
}

function isBarEnabled(series) {
    //Trash
    var helpers = DataProcessor.helpers,
        cType = helpers.getChartType(series.type || this.globalChartType, NUMERIC_DATATYPE);
    return !series.disabled && isBarChartType(cType);
}

function getBandCount(chartObj) {
    let dataObj = chartObj.dataObject;
    if (isFacet(chartObj) && dataObj.isCombinationalChart) {
        var gridList = getGridList(chartObj.dataObject);

        return d3_max(gridList, grid => {
            let { data } = grid;

            return data.reduce(function(prev, cur) {
                return prev + (isBarEnabled.call(dataObj, cur.node) ? 1 : 0);
            }, 0);
        });
    } else {
        return d3_sum(
            chartObj.seriesdata.map(function(d) {
                if (isBarEnabled.call(dataObj, d)) {
                    return d.data.length;
                }
            })
        ); // no of series
    }
}

/**
 *
 * @param chart
 * @param type
 * @param index
 * @return {{}|{nullAs: number}|*}
 */
function getAxisOptions(chart, type, index) {
    if (type === 'x') {
        return getAxisData(chart, type, index, false);
    } else if (type === 'clr') {
        return chart.systemConf.legend.colorBand;
    } else {
        if (chart.dataObject.hierarchical) {
            let axisOrder = getLeafAxisOrder(index, null, 'y', chart);
            return chart.systemConf.chart.axes.yaxis[axisOrder];
        }
        return getAxisData(chart, type, index, false);
    }
}

/**
 *
 * @param range
 * @param orient
 * @param reversed
 * @return {*}
 */
export function fixRangeDirection(range, orient, reversed) {
    let isTopBottomOrient = orient === 'bottom' || orient === 'top';
    if ((!reversed && !isTopBottomOrient) || (reversed && isTopBottomOrient)) {
        range = range.reverse();
    }
    return range;
}
function isPolarType(chartObj, type, index) {
    var cType = chartObj.chartTypes[0].name,
        // isPolar = (cType === 'dial' && type === 'y') || (cType === 'web' && type === 'x');
        isPolar = (cType === 'dial' && type === 'y') || (cType === 'web' && type === 'x' && !chartObj.dataObject.hierarchical);

    if (isPolar) {
        let { hierarchical } = chartObj.dataObject;
        if (hierarchical) {
            let nodeList = hierarchical.pivotLayout.getRootNodeList(type == 'x' ? 'column' : 'row');
            if (index >= nodeList.getLeavesCount()) {
                isPolar = false;
            }
        }
    }
    return isPolar;
}
/**
 *
 * @param chartObj
 * @param type
 * @param orient
 * @param axisReversed
 * @return {(number)[]|string|{optionsText: [string, string, string, string]}|[]|*[]|number[]}
 */
export function getRange(chartObj, type, orient, axisReversed, index) {
    var cType = charttypenames.get(getFirstAxisChartType(chartObj)),
        { plotarea, dataset } = chartObj,
        isPolarScaleType = dataset.getScaleType?.(type, index) == 'polar',
        isPolar = isPolarScaleType || (cType === 'dial' && type === 'y') || (cType === 'web' && type === 'x' && !chartObj.dataObject.hierarchical);

    if (isPolar) {
        let { hierarchical } = chartObj.dataObject;
        if (hierarchical) {
            let nodeList = hierarchical.pivotLayout.getRootNodeList(type == 'x' ? 'column' : 'row');
            if (index >= nodeList.getLeavesCount()) {
                isPolar = false;
            }
        }
    }

    if (isPolar) {
        return Registry.getComponent('axis.polar.helpers.getPolarRange')(chartObj, type);
    } else if (type === 'clr') {
        return chartObj.imagePallete || chartObj.seriesColor;
    } else {
        // return fixRangeDirection([0, getPlotSize(chartObj, topBottom ? 'width' : 'height', plotarea)], orient, axisReversed);

        return axisReversed
            ? orient === 'bottom' || orient == 'top'
                ? [getPlotSize(chartObj, 'width', plotarea), 0]
                : [0, getPlotSize(chartObj, 'height', plotarea)]
            : orient === 'bottom' || orient == 'top'
            ? [0, getPlotSize(chartObj, 'width', plotarea)]
            : [getPlotSize(chartObj, 'height', plotarea), 0];
    }
}

/**
 *
 * @param dataset
 * @param type
 * @param index
 */
export function getContinuousScale(dataset, type, index) {
    var column = dataset.getColumn(type, index),
        scaleType = column.scaleType || 'linear';

    return getD3Scale(scaleType);
}

export function getD3Scale(scaleType) {
    var scaleMode;

    if (scaleType === 'log') {
        scaleMode = d3_scaleSymlog;
    } else if (scaleType === 'sqrt') {
        scaleMode = d3_scaleSqrt;
    } else {
        scaleMode = d3_scaleLinear;
    }

    return scaleMode;
}

function getFirstAxisChartType(chartObj) {
    let axisChartType = chartObj.chartTypes[0].value;

    if (isFacet(chartObj)) {
        let gridList = getGridList(chartObj.dataObject);

        gridList.some(grid => {
            let paneid = `${grid.row}_${grid.column}`,
                chartType = getChartTypeByIndex(chartObj, paneid, 0).value;

            if (isCategoryOf(chartObj, chartType, 'axis')) {
                axisChartType = chartType;
                return true;
            }
        });
    }

    return axisChartType;
}
