// @zohocharts/zc/src/axis/tick-labels.js
//tick-labels.js
import { LabelRegistry } from '../label/label-registry';
import { axis, tempCanvasContext, scale } from '../core/main-zc-init';
import {
    defined,
    pick,
    invokeFunction,
    isUndefined,
    toArray,
    validateNullVal,
    utils,
    getValByPriority,
    isArray,
    getPropVal,
    bbox,
    arrayIncludes
} from '../utils/zcutils';
import { mathRound, mathCeil, mathAbs, mathLN10, mathLog, mathFloor, mathMin, mathMax, mathPow, NONE, NULL } from '../core/zc-constants';
import {
    isRotatedChart,
    getMargin,
    isTickInnerOrient,
    isTickAutoRotate,
    getOppositeOrient,
    getTickPadding,
    merged_axis_column_intersect,
    getAxisObject,
    checkTickInlineAlignWithInPlot
} from './axisutils';
import { TIME_DATATYPE, ORDINAL_DATATYPE } from '../core/datatype-constants';
import { appendEle, applyFont, SVGDOMParser, parseSVGElementsToArray } from '../utils/zcsvgutils';
import { date_utils } from './dateinterval';
import { d3_range, d3_select, d3_extent, d3_min } from '../core/d3-constants';
import { getContinuousScale, getOuterPaddingInPixel } from './scale';
import { getAutoTimeSubFunction, getFormattedValue, getFormattedValueByPercent } from '../formatter/formatter';
import { getPlotSize } from '../plot/plotSize';
import { getMaxTickSize, hasAutoFormat, distBetweenRotatedRect } from './tick-helpers';
import { formatTickWrapByMode } from './tickWrap';
import { getTextBound } from './textBound';
import { $ZCG } from '../core/zcglobal';
import { UNDEFINED } from '../core/zc-constants';
import { DataProcessor as P } from '../core/data-processor';
import { getAxisAreaDim, getLeafAxisOrder, isBranchAxis, hasBranchAxis } from './hierarchy-axis-utils';
// import { getLogTicks } from './log-tick-labels';
import { LOG } from '../core/scale-type-constant';
import { fixFloatingPoint } from '../utils/log-utils';
import { _hasDefaultBase } from '../core/data-processor/minMaxYValues';

/**
 *
 * @param {Object} obj
 * @returns {{top, right, bottom, left}}
 */

function getRotation(obj) {
    var alignmentMode = obj.alignMode,
        isRotateMode = obj.alignMode && obj.alignMode.indexOf('rotate') != -1,
        isOldRotateMode = defined(obj.rotation),
        rotation = 0,
        hasPriorityOrder,
        autoRotation = 45,
        hasDefinedAutoRotation;
    if (isRotateMode) {
        var rotationExtract = /\(([^)]+)\)/.exec(alignmentMode);
        rotation = rotationExtract ? parseInt(rotationExtract[1]) : 'auto';
    } else if (isOldRotateMode && !defined(obj.alignMode)) {
        rotation = obj.rotation == 'auto' ? obj.rotation : parseInt(obj.rotation);
        (alignmentMode = 'rotate'), (isRotateMode = true);
    } else if (Array.isArray(alignmentMode) && alignmentMode.length) {
        hasPriorityOrder = true;
        alignmentMode.some(d => {
            if (d.indexOf('rotate') != -1) {
                alignmentMode = 'rotate';
                isRotateMode = true;
                var rotationExtract = /\(([^)]+)\)/.exec(d);
                rotation = 'auto';
                hasDefinedAutoRotation = rotationExtract;
                autoRotation = rotationExtract ? parseInt(rotationExtract[1]) : autoRotation;
                return true;
            }
        });
    } else if (!defined(obj.alignMode)) {
        (alignmentMode = 'rotate(0)'), (isRotateMode = true);
    }
    return {
        alignmentMode: alignmentMode,
        isRotateMode: isRotateMode,
        rotation: rotation,
        hasPriorityOrder: hasPriorityOrder,
        hasDefinedAutoRotation: hasDefinedAutoRotation,
        autoRotation: autoRotation
    };
}

/**
 *
 * @param {Number} rotation in degree's
 * @param {Number} side in TRBL direction as 0 to 3
 * @returns {String}
 */

function autoLabelAlign(rotation, side) {
    var ret,
        angle = (pick(rotation, 0) - side * 90 + 720) % 360;
    if (angle > 15 && angle < 165) {
        ret = 'end';
    } else if (angle > 195 && angle < 345) {
        ret = 'start';
    } else {
        ret = 'middle';
    }
    return ret;
}

/**
 *
 * @param {Array} categories
 * @param {Number} expectedCount
 * @param {Number} interval
 * @param {Boolean} filterByIndices
 * @returns {Array}
 */

export function filterCategories(categories, expectedCount, interval, filterIndex) {
    var actualCount = categories.length;
    expectedCount = mathRound(expectedCount);
    interval = interval || mathCeil(actualCount / mathAbs(expectedCount));
    var temp = 0;
    var tickvalues = [],
        tickvalueIndices = filterIndex ? [] : null,
        tickvalueNodes = filterIndex ? [] : null;
    var isHierarchyData = filterIndex ? defined(categories[0].index) : false;
    while (temp < actualCount) {
        if (filterIndex) {
            if (isHierarchyData) {
                tickvalues.push(categories[temp].value);
                tickvalueIndices.push(categories[temp].index);
                tickvalueNodes.push(categories[temp]);
            } else {
                tickvalues.push(categories[temp]);
                tickvalueIndices.push(temp);
                tickvalueNodes.push({value: categories[temp], index: temp});
            }
        } else {
            tickvalues.push(categories[temp]);
        }
        temp = temp + interval;
    }

    if (filterIndex) {
        return {
            nodes: tickvalueNodes,
            ticks: tickvalues,
            indices: tickvalueIndices
        };
    } else {
        return tickvalues;
    }
}

/**
 *
 * @param formattedTickValues
 * @param tickConfig
 * @param columnInfo
 * @param {Chart} chart
 * @param {Number} columnIndex
 * @param {Number} levelIndex
 * @returns {{tickValues: Array, styles: Array, userRawTicks: Array}}
 */
function getCustomTickValues(formattedTickValues, tickConfig, columnInfo, chart, columnIndex, levelIndex) {
    var parentArray = [],
        parentStyles = [],
        userTickValues = [];

    var userColumnInfo = {
        data: [columnInfo]
    };
    columnInfo.columnindex = pick(columnInfo.columnindex, columnIndex);

    formattedTickValues.forEach(function(d, i) {
        var userInput = invokeFunction(tickConfig.showAs, d, tickConfig, userColumnInfo, chart, i, levelIndex),
            childArray = [],
            styles = [];

        if (isArray(userInput)) {
            parentArray.push(childArray);
            parentStyles.push(styles);
        } else {
            childArray = parentArray;
            styles = parentStyles;
        }
        userTickValues.push(userInput);

        if (defined(userInput)) {
            // if (isArray(userInput)) {
            //     userInput = userInput.join('\n');
            // }
            // var sandBox = document.createElement('div');
            // sandBox.innerHTML = userInput;

            // let userInputWrapped = `<zc-temp>${userInput}</zc-temp>`,
            // sandBox = new DOMParser().parseFromString(userInputWrapped, 'text/xml');
            let sandBox = SVGDOMParser(userInput);

            var tspan = parseSVGElementsToArray(sandBox, 'tspan'),
                images = parseSVGElementsToArray(sandBox, 'image,img');

            if (tspan.length) {
                tspan.forEach(function(t) {
                    childArray.push(t.textContent.trim());
                    let style = t.style || {},
                        attributes = t.attributes || {};
                    styles.push({
                        fontSize: style.fontSize && parseFloat(style.fontSize),
                        fontFamily: style.fontFamily,
                        fontStyle: style.fontStyle,
                        fontWeight: style.fontWeight,
                        x: attributes.x,
                        y: attributes.y,
                        dy: attributes.dy
                    });
                });
            } else if (!images.length) {
                childArray.push(userInput);
            }

            if (images.length) {
                toArray(sandBox.querySelectorAll('image,img')).forEach(function(t) {
                    let style = t.style || {},
                        attributes = t.attributes || {};
                    childArray.push({
                        width: +t.getAttribute('width'),
                        height: +t.getAttribute('height')
                    });
                    styles.push({
                        y: t.getAttribute('y'),
                        dy: t.getAttribute('dy')
                    });
                });
            }
            sandBox = null;
        }
    });

    return {
        tickValues: parentArray,
        styles: parentStyles,
        userRawTicks: userTickValues
    };
}

/**
 *
 * @param {Array} ticks formatted tick values
 * @param {Array} ticksOriginal raw tick values
 * @param {Number} fontSize font-size of the tick label
 * @param {String} fontFamily font-family of the tick label
 * @param {{fontFamily: string, fontSize: number, fontStyle: string}} styles
 * @param {Boolean} calculateAsArea
 * @returns {{actualVal: *, formattedVal: String}}
 */

export function getLongestTick(ticks, ticksOriginal, fontSize, fontFamily, styles, calculateAsArea, checkWhiteSpace) {
    /**
     * @type {CanvasRenderingContext2D} tempCanvasContext
     */
    let context = tempCanvasContext()
    context.save();
    context.font = `${fontSize}px ${fontFamily}`;

    var temp = 0,
        longest = null,
        original = null,
        style,
        hasWhiteSpace,
        index;

    styles = pick(styles, []);

    if (!defined(ticksOriginal)) {
        ticksOriginal = [];
    }

    var check = function(label, i) {
        var bound = getTextBound(label, fontSize, fontFamily, styles[i]),
            width = bound.width;

        if (calculateAsArea) {
            width = bound.width * bound.height;
        }
        if (checkWhiteSpace && !hasWhiteSpace) {
            hasWhiteSpace = label.indexOf(' ') != -1;
        }
        if (width > temp) {
            temp = width;
            longest = label;
            original = ticksOriginal[i];
            style = styles[i];
            index = i;
        }
    };
    ticks.forEach(check);
    context.restore();
    return {
        actualVal: original,
        formattedVal: longest,
        style: style,
        index: index,
        hasWhiteSpace: hasWhiteSpace
    };
}

/**
 * returns a transform string
 *
 * @param {Number} rotate in degree's
 * @param {Number} x
 * @param {Number} y
 * @returns {string}
 */
function generateTransform(rotate, x, y) {
    return `rotate(${rotate} ${x || 0} ${y || 0})`;
}

/**
 * returns the size of the tick label in horizontally or vertically
 *
 * @param {String} side "width" | "height"
 * @param {Object} dimension
 * @param {Object} margin
 * @param {Number} rotation in degree's
 * @param tickFontSize
 * @param orient
 * @returns {number}
 */

export function getTickWidth(side, dimension, margin, rotation, tickFontSize, orient) {
    var space = 0;
    if (side === 'height') {
        space = Math.abs(tickFontSize + margin.top + margin.bottom);
    } else {
        if (rotation > 0) {
            space = distBetweenRotatedRect(rotation, dimension.width, tickFontSize, orient).tickSpacing + margin.left + margin.right;
        } else {
            space = dimension.width + margin.left + margin.right;
        }
    }
    return space;
}

/**
 * returns the opposite side margin based on orien
 * @param {String} orient
 * @param {Object} margin
 */

function getOppositeMargin(orient, margin) {
    return margin[getOppositeOrient(orient)];
}

/**
 *
 * @param {String} orient
 * @param {Object} tickMargin
 * @param {Object} tickMarkConfig
 * @returns {{x: number, y: number}}
 */

function getTickMargin(orient, tickMargin, tickMarkConfig, subTickMarkConfig, tickLabelConfig, isInlineToPlot, isInline, supportSubticks) {
    var tickMarkSize = pick(tickMarkConfig.size, 6),
        margin = getTickPadding(orient, tickMargin, isInline),
        x = orient === 'left' || orient === 'right' ? 'x' : 'y',
        y = orient === 'left' || orient === 'right' ? 'y' : 'x',
        k = orient === 'top' || orient === 'left' ? -1 : 1,
        k = isInlineToPlot ? -k : k,
        prop = {};

    if (supportSubticks && (subTickMarkConfig.count > 0 || defined(subTickMarkConfig.interval))) {
        tickMarkSize = mathMax(tickMarkSize, pick(subTickMarkConfig.size, 3));
    }

    prop[y] = 0;
    prop[x] = k * (margin + tickMarkSize);

    return prop;
}

/**
 * returns a empty dimension.
 * @returns {{}}
 */
function getEmptyBBox() {
    var t = {};
    t.x = t.y = t.width = t.height = 0;
    return t;
}

function addMinmaxToTickValues(tickvalues, axisticklabeldata, chartObj, type, axisorder, orient) {
    var dataObject = chartObj.dataObject;
    var minmaxlabelsdata = axisticklabeldata.minmaxlabels || {};
    if (minmaxlabelsdata.show === true) {
        var newScale = scale.getScale(chartObj, dataObject, chartObj.plotarea, orient, axisorder, chartObj.dataset, type); //Scale with actual outerPadding
        var domain = newScale.domain();
        if (validateNullVal(minmaxlabelsdata.excludeIntervals, true)) {
            tickvalues = domain;
        } else {
            if (domain[0] !== tickvalues[0]) {
                tickvalues.splice(0, 0, domain[0]);
            }
            if (domain[1] !== tickvalues[tickvalues.length - 1]) {
                tickvalues.push(domain[1]);
            }
        }
    }
    return tickvalues;
}

export function nearestNiceMin(minDomain, remaining, magnitude, interval) {
    var arr = [1, 2, 5, 10],
        len = arr.length;
    var actualMin = minDomain + remaining / 2;
    actualMin = +actualMin.toFixed(12);
    var diff, nearestMin;
    for (var j = 0; j < len; j++) {
        var a1 = magnitude / arr[j],
            curMin;
        if (minDomain % a1) {
            curMin = minDomain - (minDomain % a1) + a1;
        } else {
            curMin = minDomain;
        }

        var curDiff = mathAbs(actualMin - curMin);
        if (curDiff < diff || diff == null) {
            diff = curDiff;
            nearestMin = curMin;
        }
    }
    if (nearestMin == null || nearestMin < minDomain) {
        var e = mathFloor(mathLog(interval) / mathLN10);
        if (e < 0) {
            nearestMin = +minDomain.toFixed(mathAbs(mathMin(0, e)));
        }
        nearestMin = minDomain;
    }
    nearestMin = nearestMin > minDomain ? nearestMin : minDomain;
    return nearestMin;
}

export function nearestTickInterval(actualInterval, magnitude, multiples, type) {
    var curDiff,
        prevDiff = null,
        curInterval,
        nearestInterval = magnitude;

    multiples = multiples
        ? multiples
        : type == 'integer'
        ? [1, 1.2, 1.5, 1.75, 1.8, 1.85, 1.9, 1.95, 2, 2.2, 2.5, 3, 3.5, 4, 4.5, 5, 6, 7, 8, 9, 10]
        : [1, 1.2, 1.5, 1.75, 1.8, 1.85, 1.9, 1.95, 2, 2.2, 2.5, 3, 3.5, 4, 4.5, 5, 6, 7, 8, 9, 10];

    for (var i = 0; i < multiples.length; i++) {
        curInterval = multiples[i] * magnitude;
        if (!type || (type == 'integer' && curInterval % 1 === 0)) {
            curDiff = mathAbs(actualInterval - curInterval);
            prevDiff = prevDiff == null ? curDiff : prevDiff;
            if (curDiff < prevDiff && curInterval <= actualInterval) {
                (prevDiff = curDiff), (nearestInterval = curInterval);
            }
        }
    }
    return nearestInterval;
}

/**
 * Axis Label Rendering
 *
 * @param {Object} axisConfig
 * @param {Object} otherInfo
 * @param {String} orient
 * @param {Object} parent
 * @constructor
 */

var AxisLabel = function(axisConfig, otherInfo, orient, parent) {
    this.axis = axisConfig;
    this.config = axisConfig.label;
    this.axisRect = otherInfo.bound;
    this.parentNode = parent.isD3Selection ? parent : d3_select(parent);
    this.orient = orient;
    this.systemConfig = otherInfo.systemConf;
    this.chart = otherInfo.chartObj;
    this.axisType = otherInfo.type;
    this.order = otherInfo.axisOrder;
    this.plotRect = otherInfo.plotRect;
    this.show = !otherInfo.hide;
    this.tickObj = otherInfo.tickObj;
    if (this.show) {
        this.add();
    } else {
        this.element = appendEle(this.parentNode, 'text', [], '.', 'class', 'axislabel'); // a empty selection
    }
};

AxisLabel.prototype = {
    add: function() {
        var label = this;
        var formattedeValue = getFormattedValue({}, label.config.text, 'ordinal', this.chart, 'axis');
        let axisAreaDim = getAxisAreaDim(this.axisType, this.order, this.orient, this.chart, this.plotRect);
        label.group = appendEle(label.parentNode, 'g', [1], '.', 'class', 'axislabelgroup');
        label.element = appendEle(label.group, 'text', [1], '.', 'class', 'axislabel');
        label.text = label.config.text;
        label.element.text(formattedeValue);
        label.rect = null;
        label.transformLabel(NULL, axisAreaDim);
    },

    remove: function() {
        var label = this;
        label.rect = null;
        label.element = appendEle(label.parentNode, 'text', [], '.', 'class', 'axislabel'); // remove it
    },

    setStyles: function() {
        var label = this,
            systemConfig = label.systemConfig;
        applyFont(label.element, [label.config, systemConfig.chart, systemConfig.canvas, $ZCG]);
    },

    getBBox: function() {
        var label = this;
        if (!defined(label.rect)) {
            if (label.element.isD3Selection && !label.element.empty() && label.parentNode.style('display') !== NONE) {
                label.rect = bbox(label.element); // calculate and cache.
            } else {
                return getEmptyBBox();
            }
        }
        return label.rect;
    },

    transformLabel: function(axisBound, plotBound, side) {
        var label = this;

        if (label.show !== true) {
            return label.remove();
        }

        if ((label.element.empty() && label.show === true) || label.text !== label.config.text) {
            return label.add();
        }
        var formattedeValue = getFormattedValue({}, label.config.text, 'ordinal', this.chart, 'axis');
        label.element.text(formattedeValue);
        label.setStyles();

        var rotations = {
                top: 0,
                right: 90,
                bottom: 0,
                left: 270
            },
            systemConfig = label.systemConfig,
            textAnchor = 'middle', // it is middle for all angle
            orient = pick(side, label.orient),
            rotate = rotations[orient],
            x,
            y,
            axisRect = pick(axisBound, this.axisRect),
            plotRect = pick(plotBound, this.plotRect),
            tickLabelConfig = label.axis.ticklabel,
            axisLabelConfig = label.config,
            tickMargin = getMargin(tickLabelConfig),
            labelMargin = getMargin(axisLabelConfig),
            fontSize = parseFloat(
                utils.getVal(getValByPriority([label.config, systemConfig.chart, systemConfig.canvas, $ZCG], 'fontSize'))
            );

        // reset BBox
        if (label.fontSize !== fontSize) {
            label.rect = null;
        } else {
            label.fontSize = fontSize;
        }

        // if axis rotated is called;
        label.element.attr('dy', '-0.25em');
        var isInline = this.tickObj.isInline,
            isInwardToPlot = this.tickObj.inlineToPlot,
            isOutwardToPlot = isInline && !isInwardToPlot,
            offset = isOutwardToPlot ? 0 : tickMargin[orient];

        switch (orient) {
            case 'top': {
                var tickLength = isInwardToPlot ? 0 : axisRect.height + offset;
                x = plotRect.width / 2;
                y = tickLength + labelMargin.bottom;
                applyTransform(rotate, x, -y);
                break;
            }

            case 'right': {
                // label.element.attr("dy", "-0.25em");
                var tickLength = isInwardToPlot ? 0 : axisRect.width + offset;
                x = plotRect.height / 2;
                y = tickLength + labelMargin.left;
                applyTransform(rotate, x, -y);
                break;
            }
            case 'bottom': {
                var tickLength = isInwardToPlot ? 0 : axisRect.height + offset;
                x = plotRect.width / 2;
                y = tickLength + labelMargin.top;

                // apply baseline
                label.element.attr('dy', '0.91em');
                applyTransform(rotate, x, y);
                break;
            }
            case 'left': {
                // label.element.attr("dy", "-0.25em");
                var tickLength = isInwardToPlot ? 0 : axisRect.width + offset;
                x = tickLength + labelMargin.right;
                y = plotRect.height / 2;
                applyTransform(rotate, -x, y, -x, y);
                break;
            }
        }

        function applyTransform(rotate, x, y, xTransform, yTransform) {
            xTransform = pick(xTransform, 0);
            yTransform = pick(yTransform, 0);

            label.element.attrs({
                x: x,
                y: y,
                transform: generateTransform(rotate, xTransform, yTransform),
                'text-anchor': textAnchor
            });
        }
    }
};

/**
 *
 * @param {String} type x or y
 * @param axisInfo
 * @param plotRect
 * @param helpers
 * @param tickCount
 * @constructor
 */

var Ticks = function(type, axisInfo, plotRect, helpers, tickCount) {
    this.type = type;
    this.config = axisInfo.ticklabel;
    this.axis = axisInfo;
    this.plotRect = plotRect;
    this.helpers = helpers;

    this.tickCount = tickCount; // for web chart

    // max size
    this.maxSize = function() {
        return getMaxTickSize(this.config, this.plotRect);
    };
    this.cache = {};
};

Ticks.prototype = {
    getScale: function() {
        var tick = this,
            chart = tick.helpers.chart,
            helpers = tick.helpers;
        var hierarchical = chart.dataObject.hierarchical;
        var axisorder = chart.dataObject.categoryHierarchy && tick.type == 'y' ? 0 : helpers.axisOrder;
        var scaleObj = scale.getScale(
            chart,
            chart.dataObject,
            tick.plotRect,
            helpers.orient,
            axisorder,
            chart.dataset,
            tick.type,
            tick.extent
        );
        if (chart.dataObject.categoryHierarchy) {
            let rowList = hierarchical.pivotLayout.getRootNodeList('row'),
                columnList = hierarchical.pivotLayout.getRootNodeList('column'),
                rowCount = rowList.getCount(),
                hasMultiCluster = rowList.hasMultiCluster(),
                leafSize = rowList.leafSize,
                axisOrder = getLeafAxisOrder(helpers.axisOrder || 0, null, 'y', chart),
                axis_data = chart.systemConf.chart.axes.yaxis[axisOrder],
                hasSplitLeavesColumns = chart.systemConf.chart.axes.xaxis.splitLeaves;
            if (hasMultiCluster && tick.type == 'y') {
                if (rowCount) {
                    var parentLeafRowScale = chart.axes.y.branches[0].scale;
                    var rangeCopy = [
                        parentLeafRowScale(helpers.axisOrder) + parentLeafRowScale.discreteWidth(helpers.axisOrder),
                        parentLeafRowScale(helpers.axisOrder)
                    ];
                    if ((chart.axes.rotated && !axis_data.reversed) || (!chart.axes.rotated && axis_data.reversed)) {
                        rangeCopy = rangeCopy.reverse();
                    }
                    return scaleObj.copy().range(rangeCopy);
                } else {
                    var rangeCopy = [helpers.axisOrder * leafSize + leafSize, helpers.axisOrder * leafSize];
                    if (chart.axes.rotated) {
                        rangeCopy = rangeCopy.reverse();
                    }
                    return scaleObj.copy().range(rangeCopy);
                }
            } else if (tick.type == 'x' && hasSplitLeavesColumns) {
                var rangeCopy = [helpers.axisOrder * columnList.leafSize, helpers.axisOrder * columnList.leafSize + columnList.leafSize];
                return scaleObj.copy().range(rangeCopy);
            }
        }
        return scaleObj;
    },

    getTicks: function(plotArea, needCount, helper) {
        var tick = this,
            helpers = pick(helper, tick.helpers),
            chart = helpers.chart,
            systemConf = chart.systemConf,
            dataObject = chart.dataObject,
            plotRect = (tick.plotRect = pick(plotArea, tick.plotRect)),
            axisObj = tick.axis,
            type = tick.type,
            userTickInterval = axisObj.interval,
            userTickCount = axisObj.tickCount,
            range,
            domain,
            scale1,
            tickInterval = axisObj.interval,
            tickCount = pick(needCount, 0),
            extent = helper.extent,
            tickValues = extent,
            categoriesList,
            filterByIndex,
            tickValueIndices,
            dontFilter,
            formattedTickValues,
            axisOrder = helpers.axisOrder,
            axisRotated = helpers.axisRotated,
            dataType = chart.dataset.getDataType(type, axisOrder),
            metaData = chart.systemConf.metadata,
            columnIndex = chart.dataset.getColumnIdx(type, axisOrder),
            columnInfo = metaData.columns[columnIndex],
            originalDataType = metaData.columns[columnIndex].datatype,
            hierachical = dataObject.hierarchical,
            isMergedAxis = type == 'y' && !hierachical ? metaData.axes.y[axisOrder].length > 1 : false,
            currentMetaData = metaData.columns[columnIndex],
            formatConfig = pick(
                isMergedAxis && type == 'y' ? merged_axis_column_intersect(axisOrder, chart) : currentMetaData[originalDataType],
                {}
            ),
            formattingSubFunction = formatConfig.subfunction,
            formattingFunction = type === 'y' && chart.dataObject.isPercentChart ? getFormattedValueByPercent : getFormattedValue,
            margin = getMargin(tick.config),
            isInline = isTickInnerOrient(tick.axis, type, chart),
            inlineToPlot = isInline ? checkTickInlineAlignWithInPlot(type, helpers.orient, systemConf) : false,
            scaleType = columnInfo.scaleType,
            isLogScale = scaleType === LOG,
            tickMargin = getTickMargin(
                helpers.orient,
                margin,
                tick.axis.tickmark,
                tick.axis.subtickmark,
                tick.axis.ticklabel,
                inlineToPlot,
                isInline,
                isLogScale
            ),
            axisSize,
            tickFontSize = parseFloat(utils.getVal(getValByPriority([tick.config, systemConf.chart, systemConf.canvas, $ZCG], 'fontSize'))),
            tickFontFamily = utils.getVal(getValByPriority([tick.config, systemConf.chart, systemConf.canvas, $ZCG], 'fontFamily')),
            rotationConfig = getRotation(tick.config),
            tickAlignmentMode = rotationConfig.alignmentMode,
            tickRotateMode = rotationConfig.isRotateMode,
            tickRotation = rotationConfig.rotation,
            tickAutoRotation = rotationConfig.autoRotation,
            hasDefinedAutoRotation = rotationConfig.hasDefinedAutoRotation,
            hasTickAlignmentPriority = rotationConfig.hasPriorityOrder,
            longestTick,
            side = isRotatedChart(type, axisRotated) ? 'height' : 'width',
            hierarchical = chart.dataObject.hierarchical,
            isCategoryHierarchy = hierarchical && hierarchical.categoryTree,
            isAutoRotation = isTickAutoRotate(chart, this.type, axisOrder, tickRotation),
            isBranchAxisType = isCategoryHierarchy ? isBranchAxis(type, axisOrder, chart) : type == 'x' && axisOrder > 0,
            rotation = isBranchAxisType ? 0 : tickRotation, //isAutoRotation || isHierarchy ? 0 : +tick.config.rotation,
            //
            firstChartName = chart.chartTypes[0].name,
            commonPlotOptions = chart.systemConf.chart.plot.plotoptions[firstChartName],
            outerPadding = dataObject.isAxisCategory ? commonPlotOptions.outerPadding || 0 : 0,
            pad = dataObject.isAxisCategory ? commonPlotOptions.padding || 0.2 : 0,
            //
            longestFormattedString = null,
            plotSize = getPlotSize(helpers.chart, side, plotRect),
            padding = type === 'x' ? plotSize * outerPadding * 2 : 0,
            dimension,
            space,
            customTickInfo = {},
            maxTickSize = helper.maxTickSize || this.maxSize(),
            hasTimeGroupBy =
                (isArray(formatConfig.groupby) && formatConfig.groupby.length > 0) ||
                (formatConfig.groupby === 'AUTO' && formattingSubFunction !== 'YEAR'),
            minorTickValues,
            axisInfo = getAxisObject(chart, type, axisOrder);

        var specifierIndex = hasTimeGroupBy ? axisOrder : null,
            existingScale,
            behaviours = getPropVal(chart, 'behaviouralEvents.eventManager.behaviours') || {},
            zoomedPlotSize = null,
            hasSmoothEffect =
                behaviours.PLOT_ZOOM || (behaviours.CHART_SCROLL && chart.scroll && chart.scroll.settings.smooth && type === 'x');

        if (hasTimeGroupBy && formatConfig.groupby === 'AUTO') {
            //#ZC1015
            formatConfig.AUTO_GROUP_ORDER = P.HierarchicalTimeDataAdapter.getAutoGroupOrder(formattingSubFunction).reverse();
        }

        if (behaviours.PLOT_ZOOM || behaviours.PLOT_BRUSH || behaviours.CHART_SCROLL) {
            // let axisInfo = getAxisObject(chart, type, axisOrder);
            existingScale = axisInfo.scale;

            // #ZC2133
            // for scroll apply only x axis
            if (type === 'x' && behaviours.CHART_SCROLL) {
                // #ZC2287 $chart.viewPortExtent first
                extent = chart.viewPortExtent || axisInfo.domain || extent;
            } else {
                extent = axisInfo.domain || extent;
            }

            tickValues = hasSmoothEffect ? null : extent;

            let newRange = existingScale.range();
            zoomedPlotSize = plotSize = mathAbs(newRange[1] - newRange[0]);
            padding = type === 'x' ? plotSize * outerPadding * 2 : 0;

            if (dataType === 'time' && isArray(extent) && existingScale) {
                if (extent[0] instanceof Date) {
                    extent = extent.map(d => d.getTime());
                }

                let diff = mathAbs(extent[1] - extent[0]),
                    correction = diff * existingScale.outerPadding;

                extent[0] = extent[0] + correction;
                extent[1] = extent[1] - correction;
            }
        }

        if (!isAutoRotation) {
            if (rotation < 0) {
                rotation = 360 - mathAbs(rotation);
            } else if (rotation >= 360) {
                rotation = mathAbs(rotation) % 360;
            }
        }
        if (tickAutoRotation) {
            if (tickAutoRotation < 0) {
                tickAutoRotation = 360 - mathAbs(tickAutoRotation);
            } else if (tickAutoRotation > 360) {
                tickAutoRotation = mathAbs(tickAutoRotation) % 360;
            }
        }
        var rowList, columnList, leavesCount, clustersCount, splitLeavesColumn, gridSize;

        if (hierarchical) {
            (rowList = hierarchical.pivotLayout.getRootNodeList('row')),
                (columnList = hierarchical.pivotLayout.getRootNodeList('column')),
                (leavesCount = rowList.getLeavesCount()),
                (clustersCount = rowList.getClustersCount());
            splitLeavesColumn = columnList.splitColumnLeaves;
        }

        switch (dataType) {
            case 'numeric': {
                var encodeOrder =
                        type == 'y' && isCategoryHierarchy ? (rowList.hasMultiCluster() ? helpers.axisOrder : 0) : helpers.axisOrder,
                    // order = type == 'y' && hierarchical ? hierarchical.leafRows.hasMulti && helpers.axisOrder <  hierarchical.leafRows.totalCount ? (helpers.axisOrder % hierarchical.leafRows.count) : 0 : helpers.axisOrder;
                    order =
                        type == 'y' && isCategoryHierarchy
                            ? rowList.hasMultiCluster() && helpers.axisOrder < leavesCount
                                ? helpers.axisOrder % clustersCount
                                : 0
                            : helpers.axisOrder;

                // clear the existing domain
                axisInfo.logDomain = null;

                domain = defined(extent)
                    ? extent
                    : type === 'x'
                    ? dataObject.getMinMaxX() //TODO need to pass UNDEFINED correction as same as minmax-y
                    : dataObject.getMinMaxY(order, UNDEFINED, chart.datalabels.hasFixedPosition);

                var correction4Datalabels;
                if (type === 'y' && chart.datalabels.hasFixedPosition) {
                    var rangeCorrections_datalabels = chart.cache.correction,
                        id = `label-${type}-${helpers.axisOrder}`,
                        labelCorrection = 0;

                    if (isUndefined(rangeCorrections_datalabels[id])) {
                        var yEncode = dataObject.getEncode('y', helpers.axisOrder);
                        labelCorrection = LabelRegistry.get('preCalculation')(type, yEncode.seriesData, chart) || 0;
                        rangeCorrections_datalabels[id] = [labelCorrection, labelCorrection];
                    } else {
                        labelCorrection = rangeCorrections_datalabels[id][1] || 0;
                    }
                    correction4Datalabels = labelCorrection;
                } else {
                    correction4Datalabels = 0;
                }
                var height;
                if (type == 'y' && isCategoryHierarchy && leavesCount) {
                    var rangeCopy = [helpers.axisOrder * rowList.leafSize + rowList.leafSize, helpers.axisOrder * rowList.leafSize];
                    gridSize = height = mathAbs(rangeCopy[1] - rangeCopy[0]);
                }
                if (type == 'x' && isCategoryHierarchy && splitLeavesColumn) {
                    var rangeCopy = [
                        helpers.axisOrder * columnList.leafSize,
                        helpers.axisOrder * columnList.leafSize + columnList.leafSize
                    ];
                    gridSize = height = mathAbs(rangeCopy[1] - rangeCopy[0]);
                }
                // define axis size and range
                if (side === 'height') {
                    axisSize = (height || plotRect.height) - correction4Datalabels;
                    tickCount = tickCount || mathRound(axisSize / 40);
                    range = [axisSize, 0];
                } else {
                    axisSize = (height || plotRect.width) - correction4Datalabels;
                    tickCount = tickCount || mathRound(axisSize / 100);
                    range = [0, axisSize];
                }

                // define tick interval
                if (defined(userTickCount)) {
                    tickInterval = (domain[1] - domain[0]) / userTickCount;
                }

                let calculatedInterval = (domain[1] - domain[0]) / userTickInterval;
                var useTickInterVal = defined(userTickInterval) && calculatedInterval < axisSize;

                // numeric scale
                // var scaleObj =
                //     existingScale ||
                //     d3_scaleLinear()
                //         .domain(domain)
                //         .range(range);
                var scaleObj =
                    existingScale ||
                    getContinuousScale(chart.dataset, type, axisOrder)()
                        .domain(domain)
                        .range(range);

                // get ticks
                if (isLogScale) {
                    const logTicks = scale.log.ticks;
                    if (scaleObj.constant) {
                        let min = d3_min(domain);
                        if (min > 0 && min < pick(axisObj.logBase, 10) && axisObj.allowNonPositiveTicks === false) {
                            scaleObj.constant(0.0001);
                        } else {
                            scaleObj.constant(1);
                        }
                    }

                    // #ZC2805 $domain changed to $dataMinMax
                    let dataMinMax = type === 'x' ? dataObject.getMinMaxX(0) : dataObject.getMinMaxY(order, 0),
                        addBaseline = !arrayIncludes(dataMinMax),
                        domainCopy = dataMinMax.slice(0),
                        logDomain,
                        usePadding = !(dataObject.isPercentChart || dataObject.hasLevelMarker || dataObject.isWaterfallEnabled),
                        adjustToBase = null,
                        yEncode = dataObject.getEncode('y', order);

                    // FEATURES_LOGAXIS_45
                    if (dataMinMax[0] === 0 && dataMinMax[1] === 0) {
                        columnInfo.scaleType = null;
                        domain =
                            type === 'x'
                                ? dataObject.getMinMaxX(0)
                                : dataObject.getMinMaxY(order, UNDEFINED, chart.datalabels.hasFixedPosition);
                        columnInfo.scaleType = LOG;
                    }

                    // type y case CXZA_LEGEND_26
                    if (
                        yEncode &&
                        _hasDefaultBase(yEncode, chart.dataObject, chart) &&
                        addBaseline &&
                        type === 'y' &&
                        !defined(axisObj.minRange)
                    ) {
                        adjustToBase = 0;
                        domainCopy.push(0);
                        domainCopy = d3_extent(domainCopy);
                    }

                    const formatFunc = tickValues => {
                        return formatArray(
                            tickValues,
                            formattingFunction,
                            formatConfig,
                            chart,
                            originalDataType,
                            'axis',
                            specifierIndex,
                            tickAlignmentMode,
                            maxTickSize,
                            tickFontSize
                        );
                    };

                    const addMinMaxLog = tickValues => {
                        axisInfo.logDomain = d3_extent(tickValues);
                        return addMinmaxToTickValues(tickValues, tick.config, chart, type, helpers.axisOrder, helpers.orient);
                    };

                    // fix all negative
                    ({ tickValues, minorTickValues, logDomain } = logTicks.getLogTicks(
                        domainCopy,
                        axisSize,
                        dataMinMax,
                        usePadding,
                        axisObj,
                        scaleObj.copy(),
                        margin,
                        {
                            fontSize: tickFontSize,
                            fontFamily: tickFontFamily
                        },
                        formatFunc,
                        helpers.orient,
                        tickRotation,
                        adjustToBase,
                        addMinMaxLog,
                        chart.datalabels.hasFixedPosition,
                        domain
                    ));

                    axisInfo.logDomain = logDomain;
                } else {
                    if (useTickInterVal) {
                        tickValues = getLinearTicksByInterval(domain, tickInterval, axisObj.maxRange);
                    } else if (+userTickCount > 0) {
                        tickValues = getLinearTicksByCount(domain, userTickCount, axisObj.maxRange, formattingSubFunction);
                    } else if (formattingSubFunction === 'integer') {
                        tickValues = getLinearTicksBySubFunction(domain, null, axisObj.maxRange);
                    } else {
                        tickValues = scaleObj.ticks(tickCount);
                    }
                }

                // add min max label if needed
                // #ZC2808
                tickValues = addMinmaxToTickValues(tickValues, tick.config, chart, type, helpers.axisOrder, helpers.orient);
                tickCount = tickValues.length; //(this.config.zigzag) ? tickValues.length/2 : tickValues.length;
                break;
            }
            case 'ordinalNonLinear':
            case 'ordinal': {
                var visibleTickCount,
                    size = tickFontSize,
                    hierarchical = dataObject.hierarchical,
                    tickNodes,
                    calculatePlotWidth;
                // #ZC1188
                //filterByIndex = (defined(hierarchical) && originalDataType == ORDINAL_DATATYPE && type == "x" && axisOrder == 0) || dataObject.isWaterfallEnabled, //TODO waterfall
                filterByIndex =
                    (defined(hierarchical) && hierarchical.categoryTree && type == 'x' && !splitLeavesColumn && axisOrder == 0) ||
                    dataObject.isWaterfallEnabled; //TODO waterfall
                // #ZC1765
                calculatePlotWidth = systemConf.chart.width === 'auto' || systemConf.chart.height === 'auto';
                dontFilter = calculatePlotWidth;

                var filterTicksIndexes = null;
                if (dataObject.isWaterfallEnabled && defined(existingScale) && isArray(tickValues)) {
                    filterTicksIndexes = tickValues;
                }

                if (!defined(tickValues)) {
                    tickValues = dataObject.getCategories(type, helpers.axisOrder).slice(0);
                }
                if (filterByIndex && defined(hierarchical)) {
                    // tickValues = dataObject.hierarchical.leaves.key.list;
                    tickValues = columnList.getClusterNodes(0);
                }

                if (originalDataType === TIME_DATATYPE && !defined(formattingSubFunction)) {
                    // set sub function
                    var info = getAutoTimeSubFunction(tickValues[tickValues.length - 1] - tickValues[0]);
                    formattingSubFunction = formatConfig.subfunction = info.subFunction;
                }

                tickCount = tickValues.length;

                dimension = {};
                dimension.width = dimension.height = tickFontSize;
                space = getTickWidth(side, dimension, margin, rotation, tickFontSize, helpers.orient);
                plotSize = splitLeavesColumn ? columnList.leafSize : plotSize;
                gridSize = splitLeavesColumn ? columnList.leafSize : null;
                visibleTickCount = mathRound((plotSize - padding) / space); // for initial filter.

                if (visibleTickCount < tickCount && side === 'width') {
                    if (rotation || tick.config.rotation > 0) {
                        space =
                            distBetweenRotatedRect(tickAutoRotation, tickFontSize, size, helpers.orient).tickSpacing +
                            margin.left +
                            margin.right;
                    }
                }
                if (!isBranchAxisType && !dontFilter) {
                    if (userTickCount < tickCount) {
                        tickValues = filterCategories(tickValues, userTickCount, null, filterByIndex);
                    } else if (defined(tickInterval)) {
                        tickValues = filterCategories(tickValues, tickCount, tickInterval, filterByIndex);
                    } else if (visibleTickCount < tickCount) {
                        tickValues = filterCategories(tickValues, visibleTickCount, null, filterByIndex);
                    }
                }

                if (filterByIndex) {
                    categoriesList = tickValues.nodes || tickValues;
                    tickValueIndices = tickValues.indices;
                    tickValues = tickValues.ticks || dataObject.getCategories(type, helpers.axisOrder, true, filterTicksIndexes).slice(0);
                }

                tickCount = tickValues.length; //(this.config.zigzag) ? tickValues.length/2 : tickValues.length;

                break;
            }

            case 'time': {
                var isAutoFormat = hasAutoFormat(formatConfig);

                if (isAutoFormat) {
                    // TODO check with axis
                    var timeScale =
                        existingScale ||
                        scale.getScale(chart, dataObject, plotRect, helpers.orient, helpers.axisOrder, chart.dataset, type, extent);
                    tickValues = timeScale.ticks();
                    formattedTickValues = tickValues.map(function(d) {
                        return timeScale.tickFormat()(d);
                    });
                    tickCount = tickValues.length;
                } else {
                    var start, end, dateInterval;
                    domain = defined(extent)
                        ? extent
                        : type === 'x'
                        ? dataObject.getMinMaxX(0)
                        : dataObject.getMinMaxY(helpers.axisOrder, 0);

                    start = new Date(domain[0]);
                    end = new Date(domain[1]);

                    // #ZC1187
                    if (!defined(currentMetaData[originalDataType])) {
                        currentMetaData[originalDataType] = formatConfig;
                    }

                    if (!defined(formattingSubFunction) || formattingSubFunction === 'AUTO') {
                        // set sub function
                        var info = getAutoTimeSubFunction(end - start);
                        formattingSubFunction = formatConfig.subfunction = info.subFunction;
                    }

                    // get the date utils 1326
                    dateInterval = new date_utils(formatConfig);
                    dateInterval.useUTC = formatConfig.useUTC === true;

                    // longestFormattedString = dateInterval.getMaxDate(formatConfig, specifierIndex); // save it
                    let maxData = dateInterval.getMaxDate(), // save it
                        longestFormattedString = getFormattedValue(formatConfig, maxData, TIME_DATATYPE, chart, 'axis', null, null, specifierIndex);
                    longestFormattedString = formatTickWrapByMode(
                        longestFormattedString,
                        tickAlignmentMode,
                        maxTickSize.horizontal,
                        maxTickSize.vertical,
                        tickFontSize
                    );
                    dimension = getTextBound(longestFormattedString, tickFontSize, tickFontFamily);
                    space = getTickWidth(side, dimension, margin, rotation, tickFontSize, helpers.orient);

                    plotSize = splitLeavesColumn ? columnList.leafSize : plotSize;
                    gridSize = splitLeavesColumn ? columnList.leafSize : plotSize;

                    tickCount = (plotSize - padding) / space;
                    // tickCount = (this.config.zigzag) ? tickCount/2 : tickCount;
                    // get the date range
                    var dateRange = dateInterval.range(start, end, mathCeil(tickCount), formattingSubFunction, tickInterval, chart);
                    // add min max tick
                    tickValues = addMinmaxToTickValues(dateRange.TICKS, tick.config, chart, type, helpers.axisOrder, helpers.orient);

                    tickCount = dateRange.TICKS.length; //(this.config.zigzag) ? dateRange.TICKS.length/2 : dateRange.TICKS.length;

                    break;
                }
            }
        }

        if (dataType === 'ordinalNonLinear') {
            var formattedTickValues = isAutoFormat
                ? formattedTickValues
                : formatArray(
                      tickValues,
                      formattingFunction,
                      formatConfig,
                      chart,
                      originalDataType,
                      'axis',
                      specifierIndex,
                      tickAlignmentMode,
                      maxTickSize,
                      tickFontSize
                  );
            // get custom tick labels
            if (defined(tick.config.showAs)) {
                customTickInfo = getCustomTickValues(tickValues, tick.config, columnInfo, chart, columnIndex, axisOrder);
                formattedTickValues = customTickInfo.tickValues;
                customTickInfo.length = formattedTickValues.length;
            }
            var longestTick = getLongestTick(formattedTickValues, tickValues, tickFontSize, tickFontFamily, customTickInfo.styles);

            // Pivot non-axis chart
            if (!chart.dataObject.isAxisCategory) {
                let _recalculateSide =
                        (type === 'x' && (axisRotated ? 'height' : 'width')) || (type === 'y' && (axisRotated ? 'width' : 'height')),
                    _toRecalculate = systemConf.chart[_recalculateSide] === 'auto';

                if (!defined(longestFormattedString)) {
                    longestFormattedString = longestTick.formattedVal;
                }

                // calculate number of tick can be fit.
                var tickBound = getTextBound(longestFormattedString, tickFontSize, tickFontFamily, longestTick.style, true);
                var labelSize = getMaxTickSize(tick.config, chart.chartarea);
                tickBound.width = mathMin(tickBound.width, labelSize.horizontal);
                tickBound.height = mathMin(tickBound.height, labelSize.vertical);

                if (_toRecalculate) {
                    tick.recalculatePlotSize(helpers.chart, plotSize, tickBound, side, tickCount, margin, pad, outerPadding, 'x');
                }

                if (_toRecalculate && hierarchical) {
                    let nodeList = hierarchical.pivotLayout.getRootNodeList('row'),
                        nodeListCount = nodeList.getCount(),
                        leavesCount = nodeList.getLeavesCount(),
                        leafSize = nodeList.leafSize;
                    if (nodeListCount) {
                        let dimension = { width: 0, height: 0 };
                        dimension[side] = leafSize;
                        tick.recalculatePlotSize(chart, plotSize, dimension, side, leavesCount, margin, pad, outerPadding, 'y');
                    }
                }
            }

            return {
                rawTickValues: tickValues,
                formatterTickValues: formattedTickValues,
                longestTickValue: longestTick,
                scale: tick.getScale(),
                longestTickDimension: getTextBound(longestTick.formattedVal, tickFontSize, tickFontFamily),
                margin: margin,
                tickMargin: tickMargin,
                rotation: rotation,
                customTickInfo: customTickInfo,
                mode: tickAlignmentMode,
                isInline: isInline,
                inlineToPlot: inlineToPlot,
                maxTickSize: maxTickSize
            };
        }

        // get a scale
        scaleObj = existingScale || tick.getScale();

        //
        var actualRange = scaleObj.actualRange;
        var isSingleTick = actualRange && actualRange[0] === actualRange[1];
        if (isSingleTick) {
            tickValues = [scaleObj.invert(actualRange[0])];

            // ZC2839
            if (isLogScale) {
                tickValues = tickValues.map(d => fixFloatingPoint(d));
            }
        }

        // chartType viewport filtering
        // #ZC2068 $hasSmoothEffect
        if (chart.renderVisiblePointsOnly && !filterByIndex) {
            tickValues = filterTickFromViewport(tickValues);
        }
        tickCount = tickValues.length;

        // format the values and get the longest tick value.
        formattedTickValues = isAutoFormat
            ? formattedTickValues
            : formatArray(
                  tickValues,
                  formattingFunction,
                  formatConfig,
                  chart,
                  originalDataType,
                  'axis',
                  specifierIndex,
                  tickAlignmentMode,
                  maxTickSize,
                  tickFontSize
              );

        // get custom tick labels
        if (defined(tick.config.showAs)) {
            customTickInfo = getCustomTickValues(tickValues, tick.config, columnInfo, chart, columnIndex, axisOrder);
            formattedTickValues = customTickInfo.tickValues;
            customTickInfo.length = formattedTickValues.length;
        }

        longestTick = getLongestTick(formattedTickValues, tickValues, tickFontSize, tickFontFamily, customTickInfo.styles);

        // uncomment the bellow condition to use the default longest date.
        if (!defined(longestFormattedString)) {
            longestFormattedString = longestTick.formattedVal;
        }

        // calculate number of tick can be fit.
        var tickBound = getTextBound(longestFormattedString, tickFontSize, tickFontFamily, longestTick.style, true);
        var labelSize = getMaxTickSize(tick.config, chart.chartarea);
        tickBound.width = mathMin(tickBound.width, labelSize.horizontal);
        tickBound.height = mathMin(tickBound.height, labelSize.vertical);
        // #ZC1765
        if (calculatePlotWidth) {
            tick.recalculatePlotSize(helpers.chart, plotSize, tickBound, side, tickCount, margin, pad, outerPadding, 'x');
        }

        if ((systemConf.chart.width === 'auto' || systemConf.chart.height === 'auto') && hierarchical && type == 'y') {
            let nodeList = hierarchical.pivotLayout.getRootNodeList('row'),
                nodeListCount = nodeList.getCount(),
                domainLength = nodeListCount ? nodeList.getDomainForBranch(0).length : 0,
                leavesCount = nodeList.getLeavesCount(),
                leafSize = nodeList.leafSize;
            if (nodeListCount && axisOrder == domainLength - 1) {
                let dimension = { width: 0, height: leafSize };
                tick.recalculatePlotSize(chart, plotSize, dimension, side, leavesCount, margin, pad, outerPadding, 'y');
            }
        }

        var bestFitTickModel = getMaxTickCount(tickCount);
        var mode = bestFitTickModel.mode,
            maxTickCount = bestFitTickModel.maxTickCount;
        rotation = bestFitTickModel.rotation;

        if (!isSingleTick && (tickCount > maxTickCount || bestFitTickModel.recalculate)) {
            // time data need a re-calculation of tick values after rotation

            if (dataType === TIME_DATATYPE) {
                if (isAutoFormat) {
                    timeScale =
                        existingScale ||
                        scale.getScale(chart, dataObject, plotRect, helpers.orient, helpers.axisOrder, chart.dataset, type, null);

                    // #ZC2211 only if
                    let reduceAutoTick = tickCount > maxTickCount && tickRotateMode && rotation === 0;

                    tickValues = timeScale.ticks(reduceAutoTick ? maxTickCount * 0.5 : maxTickCount);

                    if (tickCount > maxTickCount && !reduceAutoTick) {
                        // change the rotation manually as there are no control
                        // over calculating ticks in auto format to avoid overlapping
                        rotation = tickAutoRotation;
                    }
                    formattedTickValues = tickValues.map(d => timeScale.tickFormat()(d));
                    longestTick = getLongestTick(formattedTickValues, tickValues, tickFontSize, tickFontFamily, customTickInfo.styles);
                } else {
                    dateRange = dateInterval.range(start, end, maxTickCount, formattingSubFunction, tickInterval, chart);
                    tickValues = addMinmaxToTickValues(dateRange.TICKS, tick.config, chart, type, helpers.axisOrder, helpers.orient);
                    formattedTickValues = formatArray(
                        tickValues,
                        formattingFunction,
                        formatConfig,
                        chart,
                        originalDataType,
                        'axis',
                        specifierIndex,
                        tickAlignmentMode,
                        maxTickSize,
                        tickFontSize
                    );
                }
            } else if (maxTickCount < tickCount) {
                // filter ticks if needed
                if (filterByIndex) {
                    tickValues = filterCategories(categoriesList || tickValues, maxTickCount, null, filterByIndex);
                    (tickValueIndices = tickValues.indices), (tickValues = tickValues.ticks);
                } else {
                    tickValues = filterCategories(tickValues, maxTickCount);
                    if (tick.config.showAs) {
                        // TODO have convert multiple object to single object to reduce loop.
                        customTickInfo.styles = filterCategories(customTickInfo.styles, maxTickCount);
                        customTickInfo.userRawTicks = filterCategories(customTickInfo.userRawTicks, maxTickCount);
                    }
                }

                formattedTickValues = filterCategories(formattedTickValues, maxTickCount);
                longestTick = getLongestTick(formattedTickValues, tickValues, tickFontSize, tickFontFamily, customTickInfo.styles);
            }
        }

        // time data group by
        if (originalDataType === TIME_DATATYPE && hasTimeGroupBy) {
            if (hasTimeGroupBy && !defined(dataObject.hierarchical)) {
                dataObject.hierarchical = new P.HierarchicalTimeDataAdapter(
                    chart,
                    tickValues,
                    formatConfig.groupby,
                    dataType !== TIME_DATATYPE,
                    columnInfo
                );
            } else if (hasTimeGroupBy) {
                dataObject.hierarchical.update(tickValues);

                /*
                Find ranges for column branches
                */
                let rangeScaleUtils = dataObject.hierarchical.setScaleForRange(chart.axes.x.rangeScale);
                hierarchical.pivotLayout.setRanges('column', rangeScaleUtils);
            }
        }

        if (defined(customTickInfo.styles) || (formatConfig && formatConfig.format && formatConfig.format.breakSpace)) {
            longestTick = getLongestTick(formattedTickValues, tickValues, tickFontSize, tickFontFamily, customTickInfo.styles, true);
        }

        var textDimension = getTextBound(longestFormattedString, tickFontSize, tickFontFamily, longestTick.style);

        return {
            rawTickValues: tickValues,
            formatterTickValues: formattedTickValues,
            longestTickValue: longestTick,
            scale: scaleObj,
            longestTickDimension: textDimension,
            margin: margin,
            tickMargin: tickMargin,
            rotation: rotation,
            mode: mode,
            isInline: isInline,
            inlineToPlot: inlineToPlot,
            tickValueIndices: tickValueIndices,
            customTickInfo: customTickInfo,
            maxTickSize: maxTickSize,
            minorTickValues
        };

        /**
         *
         * @param ticks
         * @return {*}
         */
        function filterTickFromViewport(ticks) {
            // $span instead of plotarea because sometime during resize, the plotarea changes before range
            // and the tick values gets filtered out
            let span = mathAbs(scaleObj.origRange[1] - scaleObj.origRange[0]),
                correction = scaleObj.bandwidth ? scaleObj.bandwidth() / 2 : 0,
                r = [-correction, span - correction];

            return ticks.filter(t => {
                let pos = scaleObj(t);
                return pos !== null && pos >= r[0] && pos <= r[1];
            });
        }
        // now handle over flow of the formatted ticks and return;

        function getMaxTickCount(tickCount) {
            if (tickAlignmentMode === 'auto') {
                //First check for normal ticks with rotation as 0
                var maxCount_Rotate_0 = (tickBound, 0);
                //second check for zigzag only for ordinal datatype, for numeric and time datatype go for rotation with 45
                if (maxCount_Rotate_0.maxTickCount < tickCount) {
                    var maxCount_zigzag =
                        dataType == ORDINAL_DATATYPE
                            ? getMaxTickCountByZigzag(tickBound)
                            : getMaxTickCountByRotation(tickBound, tickAutoRotation, true);
                    //third for tick rotation with 45
                    if (maxCount_zigzag.maxTickCount < tickCount) {
                        var maxCount_Rotate_45 = getMaxTickCountByRotation(tickBound, tickAutoRotation, true);
                        //fourth for ticks with white space
                        if (maxCount_Rotate_45.maxTickCount < tickCount && longestTick.hasWhiteSpace) {
                            return getMaxTickCountByWrapOnSpace(tickBound);
                        } else {
                            return maxCount_Rotate_45;
                        }
                    } else {
                        return maxCount_zigzag;
                    }
                } else {
                    return maxCount_Rotate_0;
                }
            } else if (tickRotateMode) {
                return getMaxTickCountByRotation(tickBound);
            } else if (tickAlignmentMode == 'zigzag') {
                return getMaxTickCountByZigzag(tickBound);
            } else if (tickAlignmentMode == 'wrapOnSpace') {
                return getMaxTickCountByWrapOnSpace();
            } else if (tickAlignmentMode == 'wrapAll') {
                return getMaxTickCountByWrapAll();
            }
        }

        function getMaxTickCountByBounds(rotation, customBounds, customType) {
            var tickArea,
                temp,
                padding = 0,
                countForAngle,
                bounds;

            var length = getPlotSize(helpers.chart, side, tick.plotRect, gridSize);
            if (customType === 'x') {
                padding = length * outerPadding * 2;
            }

            tickArea = length - padding;
            countForAngle = !!rotation;
            bounds = customBounds || tickBound;
            if (countForAngle) {
                if (side === 'height' && (rotation === 90 || rotation === 270)) {
                    return tickArea / (bounds.width + mathAbs(margin.top) + mathAbs(margin.bottom));
                }

                var r = rotation % 90;
                var angleCheck = side === 'height' ? r > 0 : r > 0 || rotation === 90 || rotation === 270;

                if (angleCheck) {
                    var distBwTicks = distBetweenRotatedRect(rotation, bounds.width, bounds.height, helpers.orient);
                    return tickArea / (distBwTicks.tickSpacing + mathAbs(margin.left) + mathAbs(margin.right));
                }
            }

            if (side === 'height') {
                temp = bounds.height + mathAbs(margin.top) + mathAbs(margin.bottom);
            } else {
                temp = bounds.width + mathAbs(margin.left) + mathAbs(margin.right);
            }
            return tickArea / temp;
        }

        function getMaxTickCountByRotation(tickBoundgetMaxTickCountByRotation, customRotation, recalculate) {
            var rotation_new = defined(customRotation) ? customRotation : rotation,
                categoryHierarchy = chart.dataObject.categoryHierarchy;
            var count;
            if (rotation_new === 'auto') {
                rotation = 0;
                count = getMaxTickCountByBounds(rotation, tickBound, type);
                if (!isSingleTick && count < tickCount) {
                    if (side === 'width') {
                        // isBranchAxisType = chart.dataObject.pivot && type == 'x' && isBranchAxis(type, axisOrder, chart),
                        if (
                            !hasDefinedAutoRotation &&
                            ((categoryHierarchy && type == 'x' && !isBranchAxisType && hasBranchAxis(type, helpers.chart)) ||
                                (!categoryHierarchy && type == 'x' && helpers.chart.axes.x.count > 1))
                        ) {
                            rotation = 90; //For hierarchy auto rotate 90
                        } else {
                            rotation = tickAutoRotation;
                        }

                        count = getMaxTickCountByBounds(rotation, tickBound, null); //For auto rotation calculation customType should be null
                        recalculate = true;
                    } else {
                        recalculate = true;
                    }
                } else {
                    recalculate = false;
                }
            } else {
                count = getMaxTickCountByBounds(rotation_new, tickBound, type);
                rotation = rotation_new;
            }
            return {
                mode: 'rotate',
                rotation: rotation,
                maxTickCount: count,
                recalculate: recalculate
            };
        }

        function getMaxTickCountByInline(tickBound) {
            var count = getMaxTickCountByBounds(0, tickBound, type); //For zigzag rotation is 0

            return {
                mode: 'inline',
                rotation: 0,
                maxTickCount: count
            };
        }
        function getMaxTickCountByZigzag(tickBound) {
            var count = getMaxTickCountByBounds(0, tickBound, type); //For zigzag rotation is 0

            return {
                mode: 'zigzag',
                rotation: 0,
                maxTickCount: count * 2
            };
        }

        function getMaxTickCountByWrapOnSpace() {
            if (tickAlignmentMode == 'auto') {
                formattedTickValues = formattedTickValues.map(function(d, i) {
                    return d.split(' ');
                });
                var longestTick = getLongestTick(formattedTickValues, tickValues, tickFontSize, tickFontFamily, customTickInfo.styles);
                if (!defined(longestFormattedString)) {
                    longestFormattedString = longestTick.formattedVal;
                }

                // calculate number of tick can be fit.
                var tickBound = getTextBound(longestFormattedString, tickFontSize, tickFontFamily, longestTick.style, true);
            }

            var count = getMaxTickCountByBounds(rotation, tickBound, type);

            return {
                mode: 'wrapOnSpace',
                maxTickCount: count,
                rotation: rotation
            };
        }

        function getMaxTickCountByWrapAll() {
            var count = getMaxTickCountByBounds(rotation, tickBound, type);

            return {
                mode: 'wrapAll',
                maxTickCount: count,
                rotation: rotation
            };
        }
    },

    recalculatePlotSize: function(chartObj, oldPlotSize, tickSize, side, tickCount, margin, padding, outerPadding, type) {
        var categoryThickness = getPropVal(chartObj.systemConf, 'chart.scroll.categoryThickness') || tickSize[side],
            v,
            p;
        if (side === 'height') {
            categoryThickness += margin.top + margin.bottom;
            v = categoryThickness * tickCount;
            oldPlotSize = [v, 0];
        } else {
            categoryThickness += margin.left + margin.right;
            v = categoryThickness * tickCount;
            oldPlotSize = [0, v];
        }
        p = type == 'x' ? getOuterPaddingInPixel(oldPlotSize, tickCount, padding, outerPadding, categoryThickness) : 0;
        v += p;
        chartObj.systemConf.chart[side] = v;
    }
};

axis.Ticks = Ticks;
axis.AxisLabel = AxisLabel;

axis.Ticks.prototype.ticklabelFitModels = function(tickValues, maxTickCount) {
    var bestFitMode = {};
    var actualTickCount = tickValues.length;
    (bestFitMode.zigzag = {}), (bestFitMode.rotate = {});
    if (actualTickCount > maxTickCount) {
        if (actualTickCount * 2 < maxTickCount) {
            bestFitMode.type = 'zigzag';
        } else {
            bestFitMode.type = 'rotate';
        }
    } else {
        bestFitMode.type = 'default';
    }
    return bestFitMode;
};

function formatArray(array, formatter, formatConf, chart, dataType, type, specifierIndex, tickAlignmentMode, maxSize, fontSize) {
    var needTickConfigWrap = tickAlignmentMode == 'wrapOnSpace' || tickAlignmentMode == 'wrapAll';
    return array.map(function(value) {
        var formattedValue;
        formattedValue = formatter(formatConf, value, dataType, chart, type, null, null, specifierIndex);
        formattedValue = needTickConfigWrap
            ? formatTickWrapByMode(formattedValue, tickAlignmentMode, maxSize.horizontal, maxSize.vertical, fontSize)
            : formattedValue;
        return formattedValue;
    });
}

export function getLinearTicksByCount(domain, count, customStop, intervalType) {
    if (count === 1) {
        step = getLinearStep(domain, 1);
        min = Math.ceil(domain[0] / step) * step;
        return [min];
    } else {
        var tickInterval = (domain[1] - domain[0]) / count,
            min,
            max,
            remaining,
            m = count - 1,
            span = domain[1] - domain[0],
            step = span / m,
            power = mathFloor(mathLog(step) / mathLN10),
            magnitude = mathPow(10, power),
            ticks;

        magnitude = magnitude * Math.floor(step / magnitude);
        magnitude = +magnitude.toFixed(12); //To avoid floating point error (e.g):1.7000000000000002

        var nicedInterval = nearestTickInterval(tickInterval, magnitude, null, intervalType);

        if (intervalType == 'integer' && nicedInterval < 1) {
            return getLinearTicksBySubFunction(domain, count, customStop);
        }
        min = domain[0];
        max = domain[1];
        remaining = span - nicedInterval * m;

        // if (remaining < nicedInterval && Math.floor((span / m) / magnitude) > 1) {
        //     nicedInterval -= magnitude;
        //     remaining = span - (nicedInterval * m);
        // }

        var mag_remaining = mathPow(10, mathFloor(mathLog(remaining) / mathLN10) + 1);
        min = nearestNiceMin(min, remaining, mag_remaining, nicedInterval);
        min = +min.toFixed(12); //To avoid floating point error (e.g):1.7000000000000002
        max = min + nicedInterval * count;
        max = +max.toFixed(12); //Test_559,560

        max = max > domain[1] ? +domain[1] : max;
        ticks = d3_range(min, max, nicedInterval);
        if (defined(customStop)) {
            ticks = addMaxStopToTicks(ticks, nicedInterval, customStop);
        }
        return ticks;
    }
}

function getLinearStep(domain, count) {
    var m = defined(count) ? count : 10,
        span = domain[1] - domain[0],
        step = mathPow(10, mathFloor(mathLog(span / m) / mathLN10)),
        err = (m / span) * step;
    if (err <= 0.15) {
        step *= 10;
    } else if (err <= 0.35) {
        step *= 5;
    } else if (err <= 0.75) {
        step *= 2;
    }
    return step;
}

export function getLinearTicksByInterval(domain, tickInterval, customStop) {
    let count = mathFloor((domain[1] - domain[0]) / tickInterval),
        min,
        ticks;

    if (count) {
        let step = getLinearStep(domain, count);
        min = mathCeil(domain[0] / step) * step;

        if (min - domain[0] > tickInterval) {
            var s = mathPow(10, mathFloor(mathLog(tickInterval) / mathLN10)),
                q = mathFloor((min - domain[0]) / s);
            min -= q * s;
        }
    } else {
        var span = domain[1] - domain[0],
            s = mathPow(10, mathFloor(mathLog(span) / mathLN10));
        min = domain[0] - (domain[0] % s) + s;
    }

    ticks = d3_range(min, domain[1], tickInterval);
    if (defined(customStop)) {
        ticks = addMaxStopToTicks(ticks, tickInterval, customStop);
    }
    return ticks;
}

export function getLinearTicksBySubFunction(domain, count, customStop) {
    let ticks,
        interval = getNiceStep(domain, count);
    ticks = d3_range(getNiceMin(domain, 1), domain[1], interval);
    if (defined(customStop)) {
        ticks = addMaxStopToTicks(ticks, interval, customStop);
    }
    return ticks;
}

function getNiceMin(domain, minStep) {
    var step = getLinearStep(domain);
    if (defined(minStep)) {
        // Test_525.json
        step = mathMax(minStep, step);
    }
    return mathCeil(domain[0] / step) * step;
}

function getNiceStep(domain, count) {
    return mathCeil(getLinearStep(domain, count));
}

function addMaxStopToTicks(ticks, interval, stopValue) {
    if (ticks[ticks.length - 1] + interval <= stopValue) {
        ticks.push(ticks[ticks.length - 1] + interval);
    }
    return ticks;
}
