// @zohocharts/zc/src/axis/ticklabel-filter.js
/**
 *
 * @param {Event} event
 * @param {Object} info data with column info
 * @param {Chart} chart chartInstance
 * @param {Boolean} isDefault if true sets default styles else doesn't
 */
//ticklabel-filter.js
import { DataProcessor as P } from '../core/data-processor';
import { isArray, defined } from '../utils/zcutils';
import { axis } from '../core/main-zc-init';
import { NUMERIC_DATATYPE, TIME_DATATYPE } from '../core/datatype-constants';
import { isBranchAxis } from './hierarchy-axis-utils';

axis.TickLabelFilter = function(event, info, chart, isDefault, axisType) {
    info = info.data[0];

    //supports only for ORDINAL datatype
    if (info.datatype == NUMERIC_DATATYPE || info.datatype == TIME_DATATYPE) {
        return;
    }

    // create a cache to save the state
    if (!chart.cache.hierarchical) {
        chart.cache.hierarchical = {
            rows: {},
            columns: {}
        };
    }

    var hierarchical = chart.dataObject.hierarchical,
        itemIndex = info.itemIndex,
        tag = axisType == 'x' ? 'column' : 'row',
        cache = tag == 'row' ? chart.cache.hierarchical.rows : chart.cache.hierarchical.columns,
        nodeList = hierarchical.pivotLayout.getRootNodeList(tag),
        branchCount = nodeList.getCount(),
        leavesCount = nodeList.getLeavesCount(),
        axisOrder = info.axisOrder, // axisOrder starts from top to bottom
        depth = branchCount + leavesCount - axisOrder - 1, // depth starts from bottom to top ↑ (axisOrder to depth)
        isBranchAxisType = isBranchAxis(axisType, axisOrder, chart),
        data = nodeList.getNodesAtDepth(axisOrder),
        currentData = data[itemIndex],
        selectedData = nodeList.getAllChildren(currentData),
        key = chart.dataObject.pivot ? (tag == 'row' ? 'levelByRow' : 'levelByColumn') : 'level',
        oppositeKey = chart.dataObject.pivot ? (tag == 'row' ? 'levelByColumn' : 'levelByRow') : 'level',
        oppositeTag = chart.dataObject.pivot ? (tag == 'row' ? 'columns' : 'rows') : 'level',
        query = axis.TickLabelFilter.getQuery(selectedData[0], depth, key),
        query1 = chart.cache.hierarchical[oppositeTag].query;

    var cType = chart.chartTypes[0];

    if ((cType.name === 'area' || cType.name === 'line') && !isBranchAxisType) {
        return;
    }

    // to check the last tick label is clicked.
    cache.lastQuery = query;

    // apply default style
    if (isDefault) {
        var selection;
        if (query1) {
            var oldNodes = axis.TickLabelFilter.getSelectedNodes(chart, null, oppositeKey, query1);
            selection = oldNodes.selected;
        }
        var nodes = axis.TickLabelFilter.getSelectedNodes(chart, selection, key, query);
        axis.TickLabelFilter.getSelectedTickNodes(chart, nodes, axisType, axisOrder, itemIndex, query, key);
        applyDefaultStyles(chart, nodes, cache.query !== query);
    }

    // restore if same tick clicked twice
    if (cache.query === query) {
        if (isArray(cache.data)) {
            axis.TickLabelFilter.setState(cache.data, false);
        }
        // reset state to false
        cache.data = cache.query = cache.active = cache.depth = cache.itemIndex = cache.isLeaf = null; // clear it
    } else {
        // reset state to false
        if (isArray(cache.data)) {
            axis.TickLabelFilter.setState(cache.data, false);
        }
        // for current selection set to active mode
        axis.TickLabelFilter.setState(selectedData, true);
        cache.query = query;
        cache.data = selectedData;
        cache.active = true;
        cache.depth = depth;
        cache.itemIndex = itemIndex;
        cache.isLeaf = currentData.leaf;
    }
};

/**
 *
 * @param {Chart} chart
 * @param query
 * @param active
 */
function applyDefaultStyles(chart, nodes, active) {
    if (!nodes) {
        return;
    }

    if (active === true) {
        nodes.selected.series.style('opacity', 1);
        nodes.unSelected.series.style('opacity', 0.1);

        nodes.selected.labels.style('opacity', 1);
        nodes.unSelected.labels.style('opacity', 0.1);

        nodes.reset.tick.style('opacity', 1);
        nodes.unSelected.tick.style('opacity', 0.1);
        nodes.selected.tick.style('opacity', 1);
    } else {
        nodes.selected.series.style('opacity', 1);
        nodes.unSelected.series.style('opacity', 1);

        nodes.selected.labels.style('opacity', 1);
        nodes.unSelected.labels.style('opacity', 1);

        nodes.unSelected.tick.style('opacity', 1);
        nodes.selected.tick.style('opacity', 1);
    }
}
axis.TickLabelFilter.getSelectedTickNodes = function(chart, parentObj, axisType, axisOrder, tickIndex, query, selector) {
    if (!parentObj) {
        return;
    }

    var parentWrap = chart.wrap.selectAll(`.${axisType}-axis`),
        wrap = chart.wrap.selectAll(`.${axisType}-axis-${axisOrder}`);
    var tickSelected = wrap.selectAll(`g.tick:nth-child(${tickIndex + 2})`); // +2 refers for path.domain child and child index starts from 0

    parentObj.selected.tick = parentWrap.selectAll(`g.tick[${selector}^="${query}"]`).selectAll('text.ticklabel');
    parentObj.unSelected.tick = parentWrap.selectAll(`g.tick:not([${selector}^="${query}"])`).selectAll('text.ticklabel');
    parentObj.reset.tick = parentWrap.selectAll('g.tick').selectAll('text.ticklabel');

    return parentObj;
};
axis.TickLabelFilter.getSelectedNodes = function(chart, preSelection, selector, query) {
    var cType = chart.chartTypes[0],
        parent = preSelection ? preSelection.series : chart.svg.select('.seriesholder');
    var labels = preSelection ? preSelection.labels : chart.staticPaneSVG.select('g.labelholder');

    var querySelector = preSelection ? 'filter' : 'selectAll';

    var labelSelected = labels[querySelector](`text[${selector}^="${query}"],` + `path[${selector}^="${query}"]`),
        labelNotSelected = labels[querySelector](`text:not([level^="${query}"]), ` + `path:not([${selector}^="${query}"])`);

    var pointsSelected, pointsNotSelected;

    if (cType.name === 'bar' || cType.name === 'area' || cType.name === 'line') {
        pointsSelected = parent[querySelector](`path[${selector}^="${query}"]`);
        pointsNotSelected = parent[querySelector](`path:not([${selector}^="${query}"])`);
    } else if (cType.name === 'scatter' || cType.name === 'bubble') {
        pointsSelected = parent[querySelector](`g.scatter[${selector}^="${query}"]`);
        pointsNotSelected = parent[querySelector](`g.scatter:not([${selector}^="${query}"])`);
    } else {
        return;
    }

    return {
        selected: {
            labels: labelSelected,
            series: pointsSelected
        },
        unSelected: {
            labels: labelNotSelected,
            series: pointsNotSelected
        },
        reset: {}
    };
};

/**
 *
 * @param {[Array]} data array of data
 * @param {Boolean} isActive
 */
axis.TickLabelFilter.setState = function(data, isActive) {
    var state = function(point) {
        point.active = isActive;
    };
    data.forEach(state);
};

/**
 *
 * @param {Chart} chart
 * @param {Boolean} onDisable if the new series disabled.
 */
axis.TickLabelFilter.updateTickEventHandler = function(chart, onDisable, axisType) {
    var tag = axisType == 'x' ? 'column' : 'row',
        cache = chart.cache.hierarchical && chart.cache.hierarchical[tag],
        onVisible = !onDisable;

    if (defined(cache) && defined(cache.data)) {
        var hierarchical = chart.dataObject.hierarchical,
            firstPoint;
        var key = tag == 'row' ? 'levelByRow' : 'levelByColumn';

        if (cache.data) {
            // for the visible point sorting and taking first point
            var visiblePoints = cache.data.filter(function(point) {
                return !point.seriesDisabled;
            });

            // when a series is made disable
            if (onDisable) {
                if (visiblePoints.length) {
                    cache.query = axis.TickLabelFilter.getQuery(visiblePoints[0], cache.depth, key);
                } else {
                    cache.query = '-1';
                }
            }

            // when a series is made visible
            if (onVisible) {
                firstPoint = visiblePoints[0];

                var indexes = firstPoint[key]
                        .split('_')
                        .slice(0, cache.depth + 1)
                        .map(Number),
                    depth = indexes.length - 1,
                    itemIndex = indexes[indexes.length - 1],
                    columnList = hierarchical.pivotLayout.getRootNodeList('column'),
                    branchCount = columnList.getCount(),
                    leavesCount = columnList.getLeavesCount(),
                    axisOrder = branchCount + leavesCount - depth - 1, //depth to axisorder
                    parent = columnList.getNodesAtDepth(axisOrder),
                    currentData = parent[itemIndex],
                    // parent = hierarchical.getCategoriesByLevelIndices(indexes),
                    data = columnList.getAllChildren(currentData);

                cache.query = axis.TickLabelFilter.getQuery(firstPoint, cache.depth, key);
                cache.data = data;
                axis.TickLabelFilter.setState(data, true);
            }
        }
    }
};

axis.TickLabelFilter.getQuery = function(point, depth, key) {
    key = key || 'level';
    var indexes = point[key]
        .split('_')
        .slice(0, depth + 1)
        .map(Number);
    return `${indexes.slice(0, depth + 1).join('_')}_`;
};

export var _tickLabelFilter = axis.TickLabelFilter;
