// @zohocharts/zc/src/axis/ticklabel-overflow.js
//$Id$
//ticklabel-overflow.js

import { trimText, naiveEllipsis } from '../utils/zcsvgutils';
import { utils, getValByPriority, defined, getBoundingRect, pick } from '../utils/zcutils';
import { getAxisType, getAxisObject, getAxisData } from './axisutils';
import { scale } from '../core/main-zc-init';
import { mathAbs, mathMin, mathMax, mathCeil } from '../core/zc-constants';
import { getPlotSize } from '../plot/plotSize';
import { $ZCG } from '../core/zcglobal';
import { getMaxTickSize } from './tick-helpers';
import { getDefaultBoundingClientRect, getTspanBoundingClientRect } from './textBound';
import { d3_select } from '../core/d3-constants';

export function handleOverflowTicks(
    axisgroup,  
    orient,
    chartObj,
    rotation,
    overflow,
    scaleObj,
    wrap,
    axisComp,
    labelType,
    ticklabelOptions,
    index,
    options = {}
) 
{
    if (overflow == 'none') {
        return doHandleOverflow;
    }
    var plotarea = chartObj.plotarea,
        chartarea = chartObj.chartarea,
        canvasarea = chartObj.canvasarea,
        actChartObj = chartObj.actChartObj || chartObj,
        axisType = actChartObj.dataObject.isAxisCategory ? getAxisType(orient, actChartObj.axes.rotated) : 'y', // Pivot web chart: with multi y FEATURES_PIVOT_23
        hierarchical = actChartObj.dataObject.categoryHierarchy,
        hasDynamicLeafScale =
            chartObj && chartObj.dataObject && chartObj.dataObject.hierarchical && chartObj.dataObject.hierarchical.dynamicLeafScale,
        hasSplitLeavesColumn = hierarchical ? actChartObj.dataObject.hierarchical.pivotLayout.columns.splitColumnLeaves : false,
        xaxisdata = actChartObj.systemConf.chart.axes.xaxis,
        yaxisdata = actChartObj.systemConf.chart.axes.yaxis,
        axisOrder = index || 0,
        axisdata = getAxisData(actChartObj, axisType, axisOrder, !actChartObj.dataObject.isAxisCategory), // Pivot web chart: needMerge should be true
        overflowPattern = overflow == null || overflow == 'ellipsis' ? '..' : '',
        tickmarkdataOption = axisdata.tickmark.align,
        fSize = utils.getVal(
            getValByPriority([ticklabelOptions, actChartObj.systemConf.chart, actChartObj.systemConf.canvas, $ZCG], 'fontSize')
        ),
        ticks,
        ticklabelOptions = labelType == 'ticklabel' ? ticklabelOptions : {},
        tickObj = getAxisObject(actChartObj, axisType, index || 0).ticks,
        isHierarchialLevelType = tickObj ? tickObj.isHierarchialLevelType() : scaleObj.discreteWidth && scaleObj.bandwidth() == null,
        isPolar = options.type === 'polar',
        distBwTicks,
        count,
        dontUseMargin,
        percentLimit = 0.5;

    if (axisgroup == null && labelType == 'ticklabel') {
        ticks =
            axisComp.tickValues() == null
                ? scaleObj.ticks
                    ? scaleObj.ticks.apply(scaleObj, [10])
                    : scaleObj.domain()
                : axisComp.tickValues();
        var tickWrap = wrap.selectAll('.tick').data(ticks, scaleObj);
        axisgroup = tickWrap.selectAll('text.ticklabel');
    }

    //Update edge details for hierarchy data and tickmark align option as inbetween
    if (
        labelType == 'ticklabel' &&
        axisComp &&
        !isHierarchialLevelType &&
        ((tickObj && tickObj.isHierarchy() && axisType == 'x') || tickmarkdataOption == 'between')
    ) {
        dontUseMargin = true;
        ticks = ticks || axisComp.tickValues();
        distBwTicks = scale.getRangeBand(scaleObj, ticks, 0);
        var minTick = ticks[0],
            maxTick = ticks[ticks.length - 1];
        if (axisdata.reversed) {
            (minTick = ticks[ticks.length - 1]), (maxTick = ticks[0]);
        }

        var outerPadding = mathAbs(distBwTicks.outerPaddingInPixel),
            bandWidth = scaleObj.bandwidth ? scaleObj.bandwidth() / 2 : 0;
    }

    var labelSize = getMaxTickSize(ticklabelOptions, chartarea),
        hasUserLabelSize = defined(ticklabelOptions.width) || defined(ticklabelOptions.maxWidth),
        maxTickSize_Vert = labelSize.vertical,
        maxTickSize_Hori = labelSize.horizontal;

    var edge = getBoundaryEdges(actChartObj),
        margin = getExtraMarginForBoundary(labelType, plotarea, canvasarea, chartarea, percentLimit);

    let axisgroupNode = axisgroup.node();
    if (chartObj.actChartObj || (labelType == 'axislabel' && axisgroupNode)) {
        return doHandleOverflow(axisgroupNode);
    } else {
        return doHandleOverflow;
    }

    function getBoundaryEdges(chartObj, _bandwidth) {
        var _bandwidthCopy = pick(_bandwidth, bandWidth);
        var edge = {};
        var scale1 = axisType == 'x' && hasSplitLeavesColumn && chartObj.axes.x.leaves[0] ? chartObj.axes.x.leaves[0].scale : scaleObj;
        if (dontUseMargin) {
            if (orient == 'bottom' || orient == 'top') {
                (edge.left = scale1(minTick) - mathAbs(_bandwidthCopy - outerPadding)),
                    (edge.right = scale1(maxTick) + (outerPadding + _bandwidthCopy));
            } else if ((orient == 'left' || orient == 'right') && (rotation == 90 || rotation == 270)) {
                (edge.top = scale1(maxTick) - outerPadding), (edge.bottom = scale1(minTick) + outerPadding);
            }
        } else {
            edge.top = edge.left = 0;
            edge.right = orient == 'bottom' || orient == 'top' ? getPlotSize(chartObj, 'width', plotarea) : 0;
            edge.bottom = orient == 'bottom' || orient == 'top' ? 0 : getPlotSize(chartObj, 'height', plotarea);
        }

        return edge;
    }

    function getExtraMarginForBoundary(labelType, boundary, canvasBoundary, chartBoundary,marginPercent) {
        var margin = {};
        if (dontUseMargin || isPolar) {
            margin.left = margin.right = 0;
            margin.top = margin.bottom = 0;
        } else {
            (margin.left =
                labelType == 'axislabel'
                    ? marginPercent * (canvasBoundary.width - (boundary.left + boundary.width))
                    : marginPercent * boundary.left), //ZC743 If its axislabel margin right and left should be same, because always axislabel placed at center of plotarea with textanchor middle
                (margin.right = marginPercent * (canvasBoundary.width - (boundary.left + boundary.width))),
                (margin.top =
                    labelType == 'axislabel'
                        ? marginPercent * (boundary.top - chartBoundary.top)
                        : marginPercent * boundary.top),
                (margin.bottom =
                    labelType == 'axislabel'
                        ? marginPercent * (boundary.top - chartBoundary.top)
                        : marginPercent * boundary.bottom);
        }

        return margin;
    }

    function doHandleOverflow(tickNode, index) {
        if (overflow == 'none') {
            return;
        }
        /*
         * Cropping the text to maximum ticklabel size (according to bounded box of the text)
         */
        var activeNode = tickNode;
        var childNodes = [activeNode.childNodes[0]]; //If the SVG text has title element, the title element will be removed while changing this.textContent as because it modifies all the childNodes of it. So take the text element alone to trim

        // only for multiple line ticklabel, checking tspan
        var tspan = d3_select(activeNode)
            .selectAll('tspan')
            .nodes();
        if (tspan && tspan.length) {
            tspan.sort(function(a, b) {
                return b.textContent.length - a.textContent.length;
            });
            childNodes = tspan;
        }

        if (!childNodes) {
            return;
        }

        //Change maxTickSize if margin was not needed
        if (dontUseMargin) {
            count = count || wrap.selectAll('text.ticklabel').nodes().length;
            var outerPadding = mathAbs(distBwTicks.outerPaddingInPixel);
            if (orient == 'bottom' || orient == 'top') {
                if (hasUserLabelSize) {
                    maxTickSize_Hori = labelSize.horizontal;
                } else if (rotation != 90 && rotation != 270) {
                    maxTickSize_Hori = index == 0 || index == count - 1 ? outerPadding * 2 : mathAbs(distBwTicks.rangeBand) * 2; //DEMO_PIVOT_4_1 wrong count value generated till 2.21
                }
            } else if ((orient == 'left' || orient == 'right') && (rotation == 90 || rotation == 270)) {
                maxTickSize_Vert = index == 0 || index == count - 1 ? outerPadding * 2 : mathAbs(distBwTicks.rangeBand) * 2;
            }
        }

        //Iterate childnodes and apply ellipsis for each node
        for (var i = 0; i < childNodes.length; i++) {
            var txtNode = tspan && tspan.length ? childNodes[i] : activeNode,
                initial_textContent = txtNode.textContent,
                clientRect = txtNode.nodeName == 'tspan' ? getTspanBoundingClientRect : getDefaultBoundingClientRect,
                textBounds = clientRect(txtNode),
                initial_textBound = textBounds,
                trimmed,
                overflowLength = -1,
                max = {
                    // This should be changed for axis labels
                    width: maxTickSize_Hori,
                    height: maxTickSize_Vert
                };
                //Ticklabel trimming is getting affected due to minMargin value
                // minMargin = dontUseMargin || isPolar ? 0 : fSize; //if margin is very less, tick labels are trimmed unnecessarily. So set minMargin as its height


            //For hierarchy level type, check only for maxWidth
            if (isHierarchialLevelType && labelType == 'ticklabel' && !chartObj.actChartObj) {
                var rb =
                    orient == 'left' || orient == 'right'
                        ? maxTickSize_Hori
                        : Math.abs(scaleObj.discreteWidth(tickNode.__data__)) || maxTickSize_Hori;
                var maxWid = mathMin(maxTickSize_Hori, rb);
                naiveEllipsis(txtNode, txtNode, txtNode.textContent, maxWid, maxTickSize_Vert, textBounds, overflowPattern);
                return;
            }

            //Check for maxwidth and add ellipsis
            if (labelType == 'ticklabel') {
                if (maxTickSize_Hori < textBounds.width) {
                    overflowLength = (textBounds.width - maxTickSize_Hori) / textBounds.width;
                    var croppedTextLength = mathCeil((1 - overflowLength) * txtNode.textContent.length);
                    txtNode.textContent = txtNode.textContent.substring(0, croppedTextLength - 2) + overflowPattern; //TODO: for textOverflow = 'hidden', subtracting 2 letters is necessary?
                    textBounds = clientRect(txtNode); //update new textbounds
                }
                if (maxTickSize_Vert < textBounds.height) {
                    overflowLength = (textBounds.height - maxTickSize_Vert) / textBounds.height;
                    var croppedTextLength = mathCeil((1 - overflowLength) * txtNode.textContent.length);
                    txtNode.textContent = txtNode.textContent.substring(0, croppedTextLength - 2) + overflowPattern;
                    textBounds = clientRect(txtNode); //update new textbounds
                }
            }

            //calculte new boundary edges and textbounds
            var axisGroupBounds = isPolar ? { left: 0, top: 0 } : getBoundingRect(wrap.selectAll('.domain'));
            var newTextBounds = {
                left: textBounds.left - axisGroupBounds.left,
                right: textBounds.left + textBounds.width - axisGroupBounds.left,
                top: textBounds.top - axisGroupBounds.top,
                bottom: textBounds.top + textBounds.height - axisGroupBounds.top,
                width: textBounds.width,
                height: textBounds.height
            };

            if (options.edge) {
                edge = options.edge(txtNode, textBounds);
            } else if (hasDynamicLeafScale && scaleObj.bandwidth) {
                edge = getBoundaryEdges(chartObj, scaleObj.bandwidth(tickNode.__data__));
            }

            //Trim text based on boundary edges
            if ((orient == 'left' || orient == 'right') && (rotation != 0 && rotation != 180 && rotation != 360)) {
                if (
                    newTextBounds.top +  margin.top < edge.top ||
                    newTextBounds.bottom - margin.bottom > edge.bottom
                ) {
                    trimmed = true;
                    trimText(txtNode, txtNode.textContent, newTextBounds, edge, orient, overflowPattern, margin, max, txtNode);
                }
            } 
            else if ((orient == 'bottom' || orient == 'top') && txtNode.style['text-anchor'] == 'middle' && (rotation == 0 || rotation == 180 || rotation == 360)) {
                if (
                    newTextBounds.left +  margin.left < edge.left ||
                    newTextBounds.right - margin.right > edge.right
                ) {
                    trimmed = true;
                    trimText(txtNode, txtNode.textContent, newTextBounds, edge, orient, overflowPattern, margin, max, txtNode, null, true);
                }
            }
            else if ((orient == 'bottom' || orient == 'top') && (rotation != 90 && rotation != 270)) {
                if (
                    newTextBounds.left + margin.left < edge.left ||
                    newTextBounds.right - margin.right > edge.right
                ) {
                    trimmed = true;
                    trimText(txtNode, txtNode.textContent, newTextBounds, edge, orient, overflowPattern, margin, max, txtNode);
                }
            }

            if (labelType == 'ticklabel' && !trimmed){     
                var textBounds = clientRect(txtNode);
                if (textBounds.width != initial_textBound.width || textBounds.height != initial_textBound.height) {
                    naiveEllipsis(txtNode, activeNode, initial_textContent, max.width, max.height, textBounds, overflowPattern);
                }
            }

            if (txtNode.textContent == overflowPattern) {
                txtNode.textContent = chartObj.actChartObj ? '.' : ''; //hack  while casper automation DOC_CHART_61
            }
        }
        tickNode = null; // memory leak fix
    }
}
