// @zohocharts/zc/src/axis/tick-helpers.js
import { perToPx, pick } from '../utils/zcutils';
import { axis } from '../core/main-zc-init';
import { mathAbs, mathCos, mathPI, mathSin, mathTan } from '../core/zc-constants';
import { doOverlap } from '../utils/geometry-utils';

/**
 *
 * @param obj tick label object or axis label object
 * @param p plotarea
 * @returns {{vertical: *, horizontal: *}}
 */
export function getMaxTickSize(obj, p) {
    return {
        vertical: perToPx(pick(obj.height, obj.maxHeight, axis.ticks.maxTickSizeRatio_Vert), p.height),
        horizontal: perToPx(pick(obj.width, obj.maxWidth, axis.ticks.maxTickSizeRatio_Hori), p.width)
    };
}

/**
 * Check if the formatter is specified to auto
 *
 * @param {Object} formatConf
 * @returns {boolean}
 */

export function hasAutoFormat(formatConf, type = 'axis') {
    var format = formatConf != null ? formatConf[`${type}format`] || formatConf.format : null;
    return format && format.specifier === 'auto';
}

export function distBetweenRotatedRect(rotation, width, height, orient) {
    var ang1 = rotation * ((mathPI * 2) / 360);
    var ang2 = (90 - rotation) * ((mathPI * 2) / 360);

    if (orient === 'bottom' || orient === 'top') {
        return {
            tickSpacing: mathAbs(height / mathSin(ang1)) + 5,
            baseLen: 0
        };
    } else {
        var len1 = mathAbs(mathSin(ang2) * height);
        var len2 = mathAbs(mathCos(ang2) * height * mathTan(ang1));
        return {
            tickSpacing: len1 + len2 + 5,
            baseLen: mathAbs(mathCos(ang2) * width)
        };
    }
}

export function doTicksOverlaps(a, b, rotation, orient) {
    let bbox_a = a.getBoundingClientRect(),
        bbox_b = b.getBoundingClientRect();

    if (rotation) {
        let bbox = b.getBBox();
        let spacing = distBetweenRotatedRect(rotation, bbox.width, bbox.height, orient);

        if (mathAbs(bbox_a.x - bbox_b.x) <= spacing.tickSpacing - 5) {
            return true;
        } else {
            return false;
        }
    } else {
        let bbox1 = {
                x1: bbox_a.left,
                x2: bbox_a.left + bbox_a.width,
                y1: bbox_a.top,
                y2: bbox_a.top + bbox_a.height
            },
            bbox2 = {
                x1: bbox_b.left,
                x2: bbox_b.left + bbox_b.width,
                y1: bbox_b.top,
                y2: bbox_b.top + bbox_b.height
            };
        return doOverlap(bbox1, bbox2);
    }
}
