// @zohocharts/zc/src/scatter/parser/encode/scatter-encode.js
import { getPropVal } from '../../../utils/zcutils';
import { _setOpacityForTickFilterElements } from '../../../plot/svg/set-pivot-attr';
import { getFillColor, getStrokeColor } from '../../../utils/colorutils';
import { NULL } from '../../../core/zc-constants';
import { d3_select } from '../../../core/d3-constants';
import { BOTTOM_ALIGN, END_ALIGN, MIDDLE_ALIGN, START_ALIGN } from '../../../core/align-constants';
import { translateString } from '../../../utils/svg-transforms';
import { POINTTWEEN_ANIMATION, SIZING_ANIMATION } from '../../../core/animation-constants';

export let scatter_position = function () {
    let position_helpers = {};

    const KEY = this.KEY = 'scatter';

    const _this = this,
        chart = this.chart,
        { markerOptions, animationOptions, axisrotated } = this.builderC,
        { imageUrl } = markerOptions,
        { type: animationType } = animationOptions,
        racing = chart.racing && chart.racing.syncTransition,
        isBottomAlign = getPropVal(chart, 'axes.y.groups[0].prerenderComp.orient') === BOTTOM_ALIGN;;

    this.helpers.position[KEY] = {};

    // reference of common functions
    const { getInitialX, getInitialY, _getXYvalue } = this.builderC,
        getSymbolSize = this.oldRenderer.getSymbolSize;

    this.helpers.position[KEY].start = function (d, encode) {
        const _d = animationType == POINTTWEEN_ANIMATION && this.previousElementSibling ? this.previousElementSibling.__data__ : d.data;
        if (racing) {
            const xValue = _getXYvalue(_d);
            const yValue = axisrotated && isBottomAlign ? -chart.plotarea.top : chart.plotarea.height + 20;
            return translateString([xValue, yValue]);
        }
        return translateString([getInitialX(_d), getInitialY(_d)]) + ` ${imageUrl && animationType === SIZING_ANIMATION ? ' scale(0)' : ''}`;
    }

    this.helpers.position[KEY].end = function (d, encode) {
        const _d = d.data;
        const xValue = _getXYvalue(_d),
            yValue = _getXYvalue(_d, 'y'); 
        const transform = _this.helpers.positionComponent[KEY].transform(_d, d.index, { x: xValue, y: yValue });
        return translateString([transform.x, transform.y], transform.angle);
    }

    this.helpers.position[KEY].stop_1 = function (d, encode) {
        const _d = d.data;
        const transform = d3_select(this).attr('transform'),
            symbolSize = getSymbolSize(_d),
            previousSize = d3_select(this).attr('size');

        if (transform) {
            return `${transform} scale(${previousSize / symbolSize})`;
        }
    }

    this.helpers.position[KEY].exit = function (d, encode) {
        const _d = d.data;
        if (racing) {
            var xValue = _getXYvalue(_d);
            xValue = isNaN(xValue) ? null : xValue;
            if (axisrotated && isBottomAlign) {
                return translateString([xValue, -chart.plotarea.top])
            } else {
                return translateString([xValue, chart.plotarea.height + 20])
            }
        } else {
            return _this.helpers.position[KEY].start(d, encode);
        }
    }

    return position_helpers;
}


export let scatter_styles = function () {
    const KEY = 'scatter';
    
    let _this = this,
        chart = this.chart,
        data = this.data,
        dataUpdate = this.seriesdataAfterUpdate,
        builderC = this.builderC,
        fillComponent = getScatterFillColor.call(this), //TODO move this function here. avoid calling .call(this)
        strokeComponent = getStrokeColor(this.chart, this.builderC.datasetIndex, this.paneid, null, this.conf.params);


    // marker options
    const markerOptions = builderC.markerOptions,
        markerPosition = markerOptions.position;

    this.oldRenderer.fillColorFunc = fillComponent; //FEATURESEVENTS_TICKLABEL_6

    this.helpers.styleComponent[KEY] = {
        fillComponent, strokeComponent,
        hoverComponent: (d, i, defaultHoveroptions) => defaultHoveroptions
    };

    this.helpers.positionComponent[KEY] = {
        transform: (d, i, defaultTransform = { x, y, angle }) => defaultTransform
    }
    this.helpers.styles[KEY] = {
        visibility: function (d, i) {
            if (!markerPosition) {
                return NULL;
            }
            let position = i === 0 ? START_ALIGN : i === dataUpdate[d.subseriesIdx || 0].length - 1 ? END_ALIGN : MIDDLE_ALIGN;  // #ZC4641
            return markerPosition.indexOf(position) > -1 ? NULL : 'hidden';
        }
    };
}

function getScatterFillColor() {
    var fillColor = (this.oldRenderer.fillColorFunc = getFillColor(this.chart, this.scatterDatasetIndex, this.paneid,
        null, this.commonOption.scatterOrBubble ? null : true, null, null, this.conf.params))
    return fillColor;
}
