// @zohocharts/zc/src/scatter/parser/encode/scatter-morph-collector.js
import { d3_line } from '../../../core/d3-constants';
import { zcMorph } from '../../../morphing/zc-morping';
import { MarksRegistry } from '../../../plot/MarksRegistry';
import { getOffsetValue } from '../../../plot/morph/morphTools';
import { getSVGsymbol } from '../../../utils/svg-symbols';

export function scatter_morph_collectors() {
    let _this = this,
        { builderC, chart } = this,
        morphConfig = chart.systemConf.chart.plot.morph,
        morph_Obj = new zcMorph({ sampling: morphConfig.samplingDistance });

    let oldData = chart.oldSeriesInfo.get(builderC.datasetIndex),
        oldInfo = {
            seriesIdx: oldData.seriesIndex,
            color: oldData.color,
            transform: oldData.transformationInfo
        },
        offsetValue = getOffsetValue(chart, oldInfo.seriesIdx);

    const { _getXYvalue, markerOptions } = builderC,
        { symbol: symboltype, size: symbolsize } = markerOptions,
        symbolpath = getSVGsymbol(symboltype, symbolsize),
        getSymbolSize = this.oldRenderer.getSymbolSize;

    let interpolated = {};
    let line = d3_line();

    let current_data = this.helpers.filterData(this.data.data[0]); //subseriesIdx
    current_data.map(function (d, i) {
        let xValue = _getXYvalue(d),
            yValue = _getXYvalue(d, 'y');

        let fromPath = chart.oldSeriesInfo.get(oldInfo.seriesIdx).pathData[i];
        let fromPathAttribute = fromPath.getAttribute('d');
        fromPath.setAttribute('d', symbolpath(getSymbolSize(d)));
        let toPath = fromPath.cloneNode();
        fromPath.setAttribute('d', fromPathAttribute);

        interpolated[i] = morph_Obj.interpolate(
            {
                path: fromPath,
                xOffset: -xValue + offsetValue.x + oldInfo.transform[i].x,
                yOffset: -yValue + offsetValue.y + oldInfo.transform[i].y
            },
            { path: toPath, xOffset: 0, yOffset: 0 }
        );
    });
    _this.helpers.position.scatter_morphing = {
        interpolator: interpolated,
        line_d3: line
    };
    return interpolated;
}

MarksRegistry.add('scatter.morph', scatter_morph_collectors);
