// @zohocharts/zc/src/scatter/parser/encode/scatter-outer-encode.js
import { colorscale_helpers } from "../../../axis/colorscale-helpers";
import { FADE_ANIMATION, SIZINGRANDOM_ANIMATION, SIZING_ANIMATION } from "../../../core/animation-constants";
import { d3_select } from "../../../core/d3-constants";
import { NULL } from "../../../core/zc-constants";
import { effects } from "../../../effects/zceffects";
import { getDashArray, getLineCap } from "../../../utils/dashstyles";
import { getSVGsymbol } from "../../../utils/svg-symbols";
import { invokeFunction, pick } from "../../../utils/zcutils";

export let scatter_outer_position = function () {
    const KEY = 'outer',
        SUPER_KEY = 'scatter';

    const _this = this;
    this.helpers.position[KEY] = {};

    // dependency constants
    const getSymbolSize = this.oldRenderer.getSymbolSize,
        { shapescale, dataset } = this.conf.params,
        builderC = this.builderC,
        { imageUrl, symbol: symboltype } = builderC.markerOptions,
        { type: animationType } = builderC.animationOptions;

    this.helpers.position[KEY].start = function (d, encode) {
        const defaultShape = shapescale ? shapescale(d.data[dataset.shapeidx]) : symboltype;    // ZC4617
        const shape = _this.helpers.positionComponent[KEY].shape.call(this, d, d.index, defaultShape);
        const symbolSize = getSymbolSize(d.data),
            previousSize = d3_select(this.parentNode).attr('size');

        let size = imageUrl ? symbolSize : previousSize;
        size = previousSize != null ? size
            : (animationType == SIZING_ANIMATION || animationType === SIZINGRANDOM_ANIMATION) && !imageUrl
                ? 0 : symbolSize;

        size = _this.helpers.positionComponent[KEY].size.call(this, d, d.index, size);
        return { shape, size };
    }

    this.helpers.position[KEY].end = function (d, encode) {
        const defaultShape = shapescale ? shapescale(d.data[dataset.shapeidx]) : symboltype;    // ZC4617
        const shape = _this.helpers.positionComponent[KEY].shape.call(this, d, d.index, defaultShape);
        let size = getSymbolSize(d.data);
        size = _this.helpers.positionComponent[KEY].size.call(this, d, d.index, size);
        return { shape, size };
    }

    this.helpers.position[KEY].exit = this.helpers.position[KEY].start;
}

export let scatter_outer_styles = function () {
    const KEY = 'outer',
        SUPER_KEY = 'scatter';

    const chart = this.chart,
        builderC = this.builderC,
        rp = this.conf.params,
        { shapescale, dataset } = rp,
        color = builderC.color,
        defsLocation = chart.defsLocation,
        { markerOptions, animationOptions } = builderC,
        { imageUrl, size: symbolsize, symbol: symboltype } = markerOptions,
        renderer = this.oldRenderer,
        { getSymbolSize } = renderer,
        { scatterOrBubble } = this.commonOption;

    const imageConf = {
        name: imageUrl
    }
    const fillColor = this.helpers.styleComponent[SUPER_KEY].fillComponent,
        strokeColor = this.helpers.styleComponent[SUPER_KEY].strokeComponent;

    const hasColorScale = colorscale_helpers.isQuantileColorScale(chart);

    const animationType = animationOptions.type;

    const symbolpath = getSVGsymbol(symboltype, symbolsize),
        _symbolPath = d => (shapescale ? getSVGsymbol(shapescale(d.data[dataset.shapeidx]), symbolsize) : symbolpath);

    this.helpers.styleComponent[KEY] = {
        _symbolPath
    };

    this.helpers.positionComponent[KEY] = {
        shape: (d, i, defaultShape = symboltype) => defaultShape,
        size: (d, i, defaultSize = symbolsize) => defaultSize
    }

    this.helpers.styles[KEY] = {
        initialFillOpacity: animationType == FADE_ANIMATION ? 0 : markerOptions.outerFillOpacity,
        fillOpacity: pick(markerOptions.outerFillOpacity, 1),
        initialStrokeOpacity: animationType == FADE_ANIMATION ? 0 : markerOptions.outerStrokeOpacity,
        strokeOpacity: markerOptions.outerStrokeOpacity,
        strokeWidth: markerOptions.outerStrokeWidth,
        fill: function (d, i) {
            const symbolSize = getSymbolSize(d.data);
            const radius = symbolSize / 2;
            if (imageUrl) {
                imageConf.width = imageConf.height = parseFloat(symbolSize);
                imageConf.x = imageConf.y = -parseFloat(symbolSize) / 2;
                return `url(${defsLocation}#${effects.get(effects.svg_patterns, 'imageFill', imageConf, { chartObj: chart })})`;
            }

            return fillColor(d.index, d.data, null, invokeFunction(markerOptions[`${KEY}FillColor`],
                color, markerOptions, chart), null, scatterOrBubble ? [radius, radius] : null
            );
        },
        stroke: function (d, i) {
            const defaultStrokeColor = strokeColor(d.index, d.data, null);
            return strokeColor(d.index, d.data, invokeFunction(markerOptions[`${KEY}StrokeColor`], defaultStrokeColor, markerOptions, chart));
        },
        strokeDasharray: getDashArray(markerOptions[`${KEY}DashStyle`], markerOptions[`${KEY}StrokeWidth`]),
        strokeLinecap: getLineCap(markerOptions[`${KEY}LineCap`], markerOptions[`${KEY}DashStyle`]),
        visibility: hasColorScale && markerOptions.enabled === false ? 'hidden' : NULL
    }
}
