// @zohocharts/zc/src/effects/canvas-patterns.js
//$Id$

//canvas-patterns.js

import { Registry } from '../core/zc-registry';
import { validateNullVal, isNull, generateIdByPropValue, defined, isArray, getDocumentIndex } from '../utils/zcutils';
import { effects, getPatternSize } from './zceffects';
import { getColorCopy, rgba_canvas } from '../utils/colorutils';
import { d3_hsl, d3_map } from '../core/d3-constants';
import { getSVGsymbol } from '../utils/svg-symbols';
import { mathPI } from '../core/zc-constants';
import { canvasUtils } from '../utils/canvasutils';

export var canvas_patterns = function() {};
// #ZC1836
const CANVAS_PATTERN_STORE = d3_map();
const PATTERN_IMG_CACHE = d3_map();

canvas_patterns.canvasPatternFill = function(prop, clr, returnURL) {
    var pattern, image, canvas, context;
    canvas = document.createElement('canvas');
    context = canvas.getContext('2d');
    if (prop.ele) {
        image = generateCanvasPatternBySVGimage(prop, clr);
    } else {
        image = generateCanvasPatternByProp(prop, clr);
    }
    pattern = context.createPattern(image, 'repeat');
    return pattern;
};
canvas_patterns.imageFill = function(options, callback) {
    var image = fillSVGimageInCanvas(options.name, options, callback);
};

/**
 * TODO merge with effects.get
 *
 * @param url
 * @param options
 * @param callback
 */
function fillSVGimageInCanvas(url, options, callback) {
    let repeat_mode = options.repeatMode || 'repeat',
        /**
         * It returns the cache key for the pattern. It includes the dimension and the actual url.
         * Because the dimension can change for bubble charts.
         *
         * @param dimension
         * @return {string}
         */
        get_pattern_cache_key = dimension => {
            return `${url}-${repeat_mode}-${~~dimension.x}-${~~dimension.y}-${~~dimension.width}-${~~dimension.height}`;
        },
        /**
         * It returns a canvas pattern with the given loaded image and dimension.
         *
         * @param {Image} img
         * @param dimension
         * @param options
         * @return {CanvasPattern}
         */
        get_pattern_by_img = (img, dimension, options) => {
            let patternCanvas = document.createElement('canvas');
            patternCanvas.width = dimension.width;
            patternCanvas.height = dimension.height;
            patternCanvas.getContext('2d').drawImage(img, 0, 0, patternCanvas.width, patternCanvas.height);
            return options.context.createPattern(patternCanvas, repeat_mode);
        },
        /**
         *
         * It takes care of the calculation of the dimension and caching the pattern.
         *
         * @param {Image} img
         * @param options
         * @return {{pattern: *, dimension: *}}
         */
        get_cached_pattern = (img, options) => {
            let dimension = getImageSize4CanvasPattern(img, options),
                cache_key = get_pattern_cache_key(dimension),
                pattern = CANVAS_PATTERN_STORE.get(cache_key);

            /**
             * if the pattern is not found create with the given image and cache it
             */
            if (!defined(pattern)) {
                pattern = get_pattern_by_img(img, dimension, options);
                CANVAS_PATTERN_STORE.set(cache_key, pattern);
            }

            return { pattern, dimension };
        },
        cache_image = PATTERN_IMG_CACHE.get(url);

    if (defined(cache_image)) {
        /**
         * #ZC1836
         *
         * Here we are using the cached image to create the pattern.
         */
        let { pattern, dimension } = get_cached_pattern(cache_image, options);
        callback(options.context, pattern, dimension);
    } else {
        var img = new Image();
        img.onload = function() {
            // cache the image
            if (!PATTERN_IMG_CACHE.has(url)) {
                PATTERN_IMG_CACHE.set(url, img);
            }
            let { pattern, dimension } = get_cached_pattern(img, options);
            callback(options.context, pattern, dimension);
        };
        img.src = url;
    }
}

function getImageSize4CanvasPattern(imageObj, options) {
    var x = 0,
        y = 0,
        w = imageObj.naturalWidth,
        h = imageObj.naturalHeight,
        computedWidth = validateNullVal(options.width, 50),
        computedHeight = validateNullVal(options.height, 50);

    if (options.repeatMode === 'no-repeat') {
        if (h > w) {
            var remaining = computedWidth - computedWidth * (w / h);
            x = remaining / 2;
            computedWidth -= remaining;
        } else {
            var remaining = computedHeight - computedHeight * (h / w);
            y = remaining / 2;
            computedHeight -= remaining;
        }
    }

    return {
        x: x,
        y: y,
        width: computedWidth,
        height: computedHeight
    };
}

function generateCanvasPatternBySVGimage(properties, color) {
    var pattern = effects.get(effects.svg_patterns, 'svgPatternFill', properties, { returnSVGimage: true, base64: true });
    fillSVGimageInCanvas(pattern.url, pattern);
}

function generateCanvasPatternByProp(properties, color) {
    var patternID = `zp${generateIdByPropValue(properties, 'canvasFill')}`;
    var canvas = CANVAS_PATTERN_STORE.get(patternID); //get from cache store
    if (!defined(canvas)) {
        var patternCanvas = document.createElement('canvas');

        var bgClr_base = properties.bgClr || properties.bgClr_base,
            stroke_base = properties.stroke || properties.stroke_base;
        var regExp = /\{\{([a-z]+)\}\}/;

        var symbolTypes = properties.type.split(','),
            noOfSymbols = symbolTypes.length,
            isSize_function = typeof properties.size == 'function',
            isTranslate_function = properties.translate && typeof properties.translate == 'function',
            size = isSize_function ? properties.size(properties.width) : (properties.size || '').split(','),
            strokeWidth = properties.strokeWidth.split(','),
            fillColor = bgClr_base.split(','),
            stroke = stroke_base.split(','),
            fillOpacity = (properties.fillOpacity || '').split(','),
            strokeOpacity = (properties.strokeOpacity || '').split(','),
            translate = isTranslate_function
                ? properties.translate(properties.width)
                : properties.translate
                ? /\[.*?\]/g.test(properties.translate)
                    ? JSON.parse(properties.translate)
                    : properties.translate.split(',')
                : null;
        size = properties.width && !isSize_function ? getPatternSize(properties, size) : size;

        patternCanvas.width = size[0];
        patternCanvas.height = size[0];

        var patternCanvas_context = patternCanvas.getContext('2d');

        for (var j = 0; j < noOfSymbols; j++) {
            var stroke_cur = isNull(stroke[j] || stroke[0])
                    ? properties.stroke_base
                    : (stroke[j] || stroke[0]) == 'static'
                    ? properties.customStaticClr
                    : stroke[j] || stroke[0],
                fill_cur = isNull(fillColor[j] || fillColor[0])
                    ? properties.bgClr_base
                    : (fillColor[j] || fillColor[0]) == 'static'
                    ? properties.customStaticClr
                    : fillColor[j] || fillColor[0],
                trans_init = (size[j] || size[0]) / 2,
                trans_final = [trans_init, trans_init];
            stroke_cur = regExp.test(stroke_cur)
                ? getColorCopy(properties.stroke_base, 1, stroke_cur.replace(regExp, '$1'), d3_hsl)
                : stroke_cur;
            fill_cur = regExp.test(fill_cur) ? getColorCopy(properties.bgClr_base, 1, fill_cur.replace(regExp, '$1'), d3_hsl) : fill_cur;

            if (translate) {
                var subTranslate = translate[j] || translate[0];
                if (isArray(subTranslate)) {
                    trans_final = [trans_init + parseFloat(subTranslate[0]), trans_init + parseFloat(subTranslate[1])];
                } else {
                    var trans = trans_init + parseFloat(subTranslate);
                    trans_final = [trans, trans];
                }
            }

            patternCanvas_context.save();
            patternCanvas_context.translate(trans_final[0], trans_final[1]);
            patternCanvas_context.beginPath();
            patternCanvas_context.lineWidth = strokeWidth[j] || strokeWidth[0];
            patternCanvas_context.fillStyle = rgba_canvas(fill_cur, fillOpacity[j] || fillOpacity[0]);
            patternCanvas_context.strokeStyle = stroke_cur;

            if (symbolTypes[j] === 'circle') {
                patternCanvas_context.arc(0, 0, size[j] / 2, 0, mathPI * 2, false);
            } else {
                var patternPath = getSVGsymbol(symbolTypes[j])(size[j] || size[0]);
                canvasUtils.drawPath(patternCanvas_context, patternPath);
            }

            patternCanvas_context.fill();
            patternCanvas_context.stroke();
            patternCanvas_context.closePath();
            patternCanvas_context.restore();
        }
        CANVAS_PATTERN_STORE.set(patternID, patternCanvas); //save to cache store
        canvas = patternCanvas;
    }
    return canvas;
}

effects.canvas_patterns = canvas_patterns;
Registry.setComponent('patterns.canvas', canvas_patterns);
