// @zohocharts/zc/src/effects/gradient-helpers.js
//$Id$

import { charttype } from '../core/main-zc-init';
import { getChartTypeByIndex, isAreaChartType } from '../core/charttype-helpers';
import { $Browser } from '../core/browser-detect';
import { mathMin, mathMax, mathAbs } from '../core/zc-constants';
import { getColorCopy } from '../utils/colorutils';
import { d3_hsl } from '../core/d3-constants';
import { invokeFunction, isNull } from '../utils/zcutils';
import { Registry } from '../core/zc-registry';

export var gradient_helpers = {
    getDimension: function(chartObj, seriesIndex, paneid) {
        if (chartObj.dataObject.isPolarAxisCategory) {
            var r = chartObj.plotarea_approx.radius;
            return {
                r: r,
                x1: -r,
                x2: r,
                y1: -r,
                y2: r
            };
        } else if (chartObj._isAxisCategory()) {
            var rp = chartObj.rendererConf[seriesIndex].params,
                cType_Obj = getChartTypeByIndex(chartObj, paneid, seriesIndex),
                cType = cType_Obj.value,
                axes = chartObj.systemConf.chart.axes,
                xReversed = axes.xaxis.reversed,
                yReversed = axes.yaxis[rp.yaxiscolumnorder].reversed,
                xScale = rp.xscale,
                yScale = rp.yscale,
                xRange = xScale.range(),
                yRange = yScale.range(),
                getAdjustedRange = function(scaleObj, range) {
                    if (scaleObj.bandwidth && (isAreaChartType(cType) || cType === charttype.line)) {
                        //Test_900.json filter Oct and Nov, the line gradient should be rendered correctly
                        var bwBy2 = scaleObj.bandwidth() / 2;
                        var sign = range[0] < range[1] ? 1 : -1;
                        return [range[0] + bwBy2 * sign, range[1] - bwBy2 * sign];
                    }
                    return range;
                };

            xRange = getAdjustedRange(xScale, xRange);
            yRange = getAdjustedRange(yScale, yRange);

            if (rp.axisrotated) {
                return {
                    x1: yRange[+!yReversed],
                    x2: yRange[+!!yReversed],
                    y1: xRange[+!xReversed],
                    y2: xRange[+!!xReversed]
                };
            } else {
                return {
                    x1: xRange[+!!xReversed],
                    x2: xRange[+!xReversed],
                    y1: yRange[+!!yReversed],
                    y2: yRange[+!yReversed]
                };
            }
        }
    }
};
Registry.setComponent('gradient_helpers.getDimension', gradient_helpers.getDimension);

export function getStops(gradient, properties, color, type) {
    'use strict';
    var stopoffset = properties.stopOffset,
        stopopacity = properties.stopOpacity,
        stopcolor = properties.stopColor,
        colorgamma = properties.colorGamma || [1, 0],
        spread = properties.spreadMethod,
        radius =
            type == 'radial'
                ? +properties.radius
                : mathMax(mathAbs((+properties.x2 || 0) - (+properties.x1 || 0)), mathAbs((+properties.y2 || 0) - (+properties.y1 || 0)));
    stopoffset = stopoffset || ['0%', '100%'];
    stopcolor =
        invokeFunction(stopcolor, color) ||
        stopoffset.map(function(d, i) {
            var clr = getColorCopy(color, colorgamma[i] || 0, 'brighter', d3_hsl);
            if(isNull(clr)){
                return color;
            }
            return getColorCopy(color, colorgamma[i] || 0, 'brighter', d3_hsl);
        });
    stopopacity = stopopacity || [1, 1];

    if (!supportsSpreadMethod(spread)) {
        var newStopoffset = [],
            newStopColor = [],
            newStopOpacity = [],
            step = 0,
            repeat = 0;
        while (step < 100) {
            for (var i = 0, len = stopoffset.length; i < len; i++) {
                var newOffset = mathMin((parseFloat(stopoffset[i]) / 100) * radius + step, 100);
                newStopoffset.push(`${newOffset}%`);
                newStopOpacity.push(repeat % 2 != 0 ? stopopacity[stopopacity.length - 1 - i] : stopopacity[i]);
                newStopColor.push(repeat % 2 != 0 ? stopcolor[stopcolor.length - 1 - i] : stopcolor[i]);
            }
            step += radius;
            repeat += spread === 'reflect' ? 1 : 0;
        }
        stopoffset = newStopoffset;
        stopcolor = newStopColor;
        stopopacity = newStopOpacity;
    }
    stopcolor.map((d, i)=>{
        if(d === null){
            stopcolor[i] = color;
        }
    })
    return {
        offset: stopoffset,
        color: stopcolor,
        opacity: stopopacity
    };
}

export function supportsSpreadMethod(spreadMethod) {
    return !((spreadMethod === 'repeat' || spreadMethod === 'reflect') && ($Browser.IS_IOS || $Browser.SAFARI));
}
