// @zohocharts/zc/src/effects/patternThemes.js
//$Id$
//patternThemes.js
import { Registry } from '../core/zc-registry';
import { TRANSPARENT } from '../core/zc-constants';

export var patternutil = function() {};
export var patternThemes = {};

export var addPattern = function(name, type, size, strokeWidth, strokeClr, bgClr, fillOpacity, strokeOpacity, translate) {
    patternThemes[name] = {
        type: type || 'diagonal-line-right',
        size: size || 5,
        strokeWidth: strokeWidth || 1,
        bgClr: bgClr,
        stroke: strokeClr,
        fillOpacity: fillOpacity || 1,
        strokeOpacity: strokeOpacity || 1,
        translate: translate
    };
};
export var getPattern = function(name) {
    return {
        type: patternThemes[name].type,
        size: patternThemes[name].size,
        strokeWidth: patternThemes[name].strokeWidth,
        bgClr: patternThemes[name].bgClr,
        stroke: patternThemes[name].stroke,
        fillOpacity: patternThemes[name].fillOpacity,
        strokeOpacity: patternThemes[name].strokeOpacity,
        translate: patternThemes[name].translate
    };
};
Registry.setComponent('patterns.get', getPattern);

export function generateDefaultPatternTheme() {
    addPattern('hatch', 'diagonal-line-right', '5', '1', 'null', 'white', '1', '1');
    addPattern(
        'hatchDash',
        'square,diagonal-line-right',
        '5,5',
        '1',
        'transparent,{{brighter}}',
        'transparent,transparent',
        '1',
        '1',
        '0,3'
    );
    addPattern(
        'thinThickHatch',
        'square,diagonal-line-right,diagonal-line-right',
        '10,10,10',
        '1,1,3',
        'transparent,null,null',
        TRANSPARENT,
        '1',
        '1',
        '0,2.5,0'
    );
    addPattern(
        'thinThickHatchLeft',
        'square,diagonal-line-left,diagonal-line-left',
        '10,10,10',
        '1,1,3',
        'transparent,null,null',
        TRANSPARENT,
        '1',
        '1',
        '[0,[2.5,-2.5],0]'
    );
    addPattern('thickHatch', 'diagonal-line-right', '5', '3', 'null', 'white', '1', '1');
    addPattern('thinHatch', 'diagonal-line-right', '2', '1', 'null', 'white', '1', '1');
    addPattern('hatchLeft', 'diagonal-line-left', '5', '1', 'null', 'white', '1', '1');
    addPattern('horizontalStripes', 'horizontal-line', '5', '1', 'null', 'white', '1', '1');
    addPattern('thinHorizontalStripes', 'horizontal-line', '2', '1', 'null', 'white', '1', '1');
    addPattern('verticalStripes', 'vertical-line', '5', '1', 'null', 'white', '1', '1');
    addPattern('thinVerticalStripes', 'vertical-line', '2', '1', 'null', 'white', '1', '1');
    addPattern('crossStripes', 'vertical-line,horizontal-line', '5', '1', 'null', 'white', '1', '1');

    addPattern('tinyHatchfill', 'square,diagonal-line-right', '5', '1', 'transparent,{{brighter}}', 'null,transparent', '1', '1');
    addPattern('hatchfill', 'square,diagonal-line-right', '10', '1', 'transparent,{{brighter}}', 'null,transparent', '1', '1');

    addPattern('hatchLeftfill', 'square,diagonal-line-left', '10', '1', 'transparent,{{brighter}}', 'null,transparent', '1', '1');
    addPattern('hatchDashfill', 'square,diagonal-line-right', '5,5', '1', 'transparent,{{brighter}}', 'null,transparent', '1', '1', '0,3');
    addPattern('tinyHatchLeftfill', 'square,diagonal-line-left', '5', '1', 'transparent,{{brighter}}', 'null,transparent', '1', '1');
    addPattern('horizontalStripesfill', 'square,horizontal-line', '5', '1', 'transparent,{{brighter}}', 'null,transparent', '1', '1');
    addPattern('verticalStripesfill', 'square,vertical-line', '5', '1', 'transparent,{{brighter}}', 'null,transparent', '1', '1');
    addPattern('crossStripesfill', 'square,square', '5,5', '1', 'transparent,{{brighter}}', 'null,transparent', '1', '1');
    addPattern('checkerBoard', 'square,square,square', '10,5,5', '1', TRANSPARENT, 'transparent,null,{{brighter}}', '1', '1', '0,0,5');

    addPattern('circle', 'circle', '10', '1', 'null', 'white', '1', '1');
    addPattern('tinyCircle', 'circle', '5', '1', 'null', 'white', '1', '1');
    addPattern('crossCircle', 'square,circle,circle', '10,5,5', '1', 'transparent,null,{{brighter}}', TRANSPARENT, '1', '1', '0,0,5');
    addPattern('floatingCircle', 'square,circle', '10,5', '1', 'transparent,{{brighter}}', TRANSPARENT, '1', '1', '0,2.5');
    addPattern('dot', 'square,circle', '5,2.5', '1', 'transparent,{{brighter}}', 'transparent,transparent', '1', '1', '0,1.25');
    addPattern('dotSquare', 'square,circle', '10,2.5', '1', 'null,{{brighter}}', 'transparent,transparent', '1', '1', '0,3.75');

    addPattern('circlefill', 'square,circle', '10,10', '1', 'transparent,{{brighter}}', 'null,transparent', '1', '1');
    addPattern(
        'crossCirclefill',
        'square,circle,circle',
        '10,5,5',
        '1',
        'transparent,null,{{brighter}}',
        'null,{{brighter}},null',
        '1',
        '1',
        '0,0,5'
    );
    addPattern('floatingCirclefill', 'square,circle', '10,5', '1', 'transparent,{{brighter}}', 'null,transparent', '1', '1', '0,2.5');
    addPattern('dotfill', 'square,circle', '5,2.5', '1', 'transparent,{{brighter}}', 'null,transparent', '1', '1', '0,1.25');

    addPattern('square', 'square', '5', '1', 'null', 'white', '1', '1');
    addPattern('squarefill', 'square', '5', '1', '{{brighter}}', 'null', '1', '1');

    addPattern('star', 'star', '10', '1', 'null', 'white', '1', '1');
    addPattern('starfill', 'square,star', '10,5', '1', 'transparent,{{brighter}}', 'null,transparent', '1', '1');

    addPattern('wave', 'waves', '5', '1', 'null', 'white', '1', '1');
    addPattern('wavefill', 'square,waves', '10,10', '1', 'transparent,{{brighter}}', 'null,transparent', '1', '1');

    addPattern('cap', 'caps', '5', '1', 'null', 'white', '1', '1');
    addPattern('capfill', 'square,caps', '10,10', '1', 'transparent,{{brighter}}', 'null,transparent', '1', '1');

    addPattern('nylon', 'nylon', '5', '1', 'null', 'white', '1', '1');
    addPattern('nylonfill', 'square,nylon', '10,10', '1', 'transparent,{{brighter}}', 'null,transparent', '1', '1');
}

export function addPatternThemeByEle(ele) {
    return { ele: ele };
}
