// @zohocharts/zc/src/effects/zceffects.js
//$Id$
//zceffects.js
import { Registry } from '../core/zc-registry';
import { generateIdByPropValue, getDocumentIndex, pick, defined, simpleClone } from '../utils/zcutils';
import { d3_select } from '../core/d3-constants';
import { NONE, mathMax } from '../core/zc-constants';
import { DOCUMENT_OBJECT_HOLDER } from '../core/zc-constants';
import { addPatternThemeByEle, patternThemes, getPattern } from './patternThemes';
import { $ZCG } from '../core/zcglobal';
import { charttype } from '../core/main-zc-init';

export var effects = function() {};
Registry.setComponent('effects', effects);

effects.get = function(module, submodule, properties, args) {
    args = args || {};
    var offset = args.offset,
        dimension = args.dimension,
        returnSVGimage = args.returnSVGimage,
        base64 = args.base64,
        chartObj = args.chartObj,
        defaultModule = args.moduleType == 'svgPattern' ? effects.svg_patterns : null;
    module = module || defaultModule;
    if (!((submodule != null || submodule != '') && module[submodule])) {
        return;
    }

    var renderToSvg;

    if (chartObj && chartObj.getPatternSvg) {
        renderToSvg = chartObj.getPatternSvg();
    } else {
        renderToSvg = pick(args.renderTo, $ZCG.patternSVG);
        if (!renderToSvg.isD3Selection) {
            renderToSvg = d3_select(renderToSvg);
        }
    }

    /* Generate PatternID and check the pattern if already generated with the PatternID
     * If already exist return the patternID to apply
     * else create new pattern and return patternID.
     * */
    var effectID = `zp${generateIdByPropValue(properties, submodule)}`,
        effectEle;

    if (effectID == null) {
        return;
    }

    var documentIndex = getDocumentIndex(renderToSvg.node()),
        EFFECTS_STORE = DOCUMENT_OBJECT_HOLDER && (documentIndex) && DOCUMENT_OBJECT_HOLDER[documentIndex].EFFECTS;

    if (EFFECTS_STORE && defined(EFFECTS_STORE[effectID])) {
        effectEle = EFFECTS_STORE[effectID];
        if (returnSVGimage) {
            return effectEle;
        }
        return effectEle.id || effectEle;
    }

    effectEle = module[submodule](properties, effectID, offset, returnSVGimage, base64, chartObj, renderToSvg);
    if (effectEle) {
        if (EFFECTS_STORE) {
            EFFECTS_STORE[effectID] = effectEle;
        }
        if (returnSVGimage) {
            return effectEle;
        }
        return effectEle.id || effectEle; //for filters returns id, for patterns returns element. Need to make it common
    }
};

export function getChartEffects(chartObj) {
    var systemconf = chartObj.systemConf,
        effectName = systemconf.chart.effects.choice;
    var effect = effects.svg_filters && effects.svg_filters[effectName];
    if (!effect) {
        return;
    }
    if (chartObj.chartTypes[0].value === charttype.wordcloud) {
        return;
    }

    var properties = systemconf.chart.effects.options[effectName];
    if (effectName === 'dropShadow') {
        properties.blur = mathMax(0, properties.blur);
    }
    var effectID = effects.get(effects.svg_filters, effectName, properties, { chartObj: chartObj });
    return effectID;
}

export function getGradientProp(gradient, dimension, color, opacity) {
    var type = gradient.type;
    if (!type || type === NONE) {
        return;
    }
    var options = gradient.options[type] || gradient.options;
    var prop = simpleClone(options);
    prop.color = color != null ? color : prop.color;
    prop.opacity = opacity != null ? opacity : prop.opacity;
    prop.dimension = dimension != null ? dimension : prop.dimension;
    prop.type = type;
    // prop.defsLocation = chartObj ? chartObj.defsLocation : "";
    return prop;
}

export function getPatternProp(pattern, baseClr, opacity, customPatternDim, customClr) {
    var prop = patternThemes[pattern] ? getPattern(pattern) : addPatternThemeByEle(pattern);
    prop.bgClr_base = baseClr;
    prop.stroke_base = baseClr;
    prop.fillOpacity = prop.fillOpacity || opacity;
    prop.customStaticClr = customClr;
    if (customPatternDim) {
        (prop.x = customPatternDim.x),
            (prop.y = customPatternDim.y),
            (prop.width = customPatternDim.width),
            (prop.height = customPatternDim.height);
    }
    return prop;
}
export function getPatternSize(properties, size) {
    var patternSize = size.map(function(d, i) {
        var s = parseInt(d);
        var rem = properties.width % s;
        var quo = parseInt(properties.width / s) || 1; //if pattern size is greater than properties.width
        var partial = rem / quo;
        return partial + s;
    });
    return patternSize;
}

// export function getPatternSize(properties, size) {
//     if(properties.width == size[0]){
//         return size;
//     }
//     var patternSize = size.map(function(d, i) {
//         if(i == 0){
//             return properties.width;
//         }
//         var s = parseFloat(d);
//         var quo = (size[0] / s) || 1; //if pattern size is greater than properties.width
//         return properties.width / quo;
//     });
//     return patternSize;
// }

export function getPatternTranslate(properties, size, translate) {
    var patternSize = size[0];
    if (properties.width == patternSize) {
        return translate;
    }
    var patternSize = translate.map(function(d, i) {
        if (d == 0) {
            return 0;
        }
        var s = parseFloat(d);
        var quo = patternSize / s || 1; //if pattern size is greater than properties.width
        return properties.width / quo;
    });
    return patternSize;
}
export function applyEffect(element, options) {
    let { effect, config = {}, chartObj } = options;
    if (!effect && chartObj) {
        effect = getChartEffect(config, chartObj);
    }

    element
        .selectAll(
            '*:not(text):not(span):not(b):not(strong):not(i):not(em):not(mark):not(small):not(del):not(ins):not(sub):not(sup):not(.expandable):not(.zc-expandable):not(.minorTicks):not(.domain):not(.zc-connector-line)'
        )
        .filter(function() {
            return !this.childElementCount;
        })
        .style('filter', effect);
}

export function getChartEffect(config, chartObj) {
    let effect = config.useChartEffect !== false && chartObj.filterEffectsID;
    return effect ? `url(${chartObj.defsLocation}#${effect})` : null;
}
