// @zohocharts/zc/src/legend/events/visual-effects/size-list/highlight-item-helpers.js
//
import { $ZCG } from '../../../../core/zcglobal';
import { markerDataURI, markerBackground } from '../../../../view/list/list-helpers';
import { d3_select } from '../../../../core/d3-constants';
import { defined, pick } from '../../../../utils/zcutils';
import { applyFont } from '../../../../utils/zcsvgutils';
import { getSVGSymbolString } from '../../../../utils/svg-symbols';

export function getHighlightElement(trackerElement, i) {
    return d3_select(trackerElement.nodes()[i]);
}

export function highlightLI(ele, config) {
    let { backgroundColor, borderRadius } = config;
    ele.styles({
        'background-color': backgroundColor,
        'border-radius': defined(borderRadius) ? `${borderRadius}px` : null
    });
}

export function highlightLabel(ele, config, globalFont, priorityConfig) {
    applyFont(ele, [priorityConfig, config.list.label, globalFont, $ZCG], true);
}

export function highlightMarker(ele, scale, config, priorityConfig) {
    if (priorityConfig) {
        let background = d => {
            let width = scale.width ? scale.width(d) : scale(d),
                height = scale.height ? scale.height(d) : scale(d),
                dataURI1 = markerDataURI(config, scale)(d),
                dataURI2 = priorityConfig ? markerDataURI(priorityConfig, scale)(d) : '';

            return getSVGSymbolString(dataURI1 + dataURI2, width, height);
        };

        ele.styles({
            background,
            border: 'none',
            'border-radius': null
        });
    } else {
        ele.styles(markerBackground(config, scale));
    }
}

export function getListIndex(data, options) {
    let { config, ticks } = options,
        { reversed } = config,
        point = pick(data.point, data);

    return reversed ? ticks.length - 1 - point : point;
}
