// @zohocharts/zc/src/heatmap/heatmapchart.js
//$Id$
//heatmapchart.js
import { svg_plot_renderer } from '../core/main-zc-init';
import { getAnimationConf, cancelAnimation, fadeTween, flipTween, scaleTween } from '../utils/animationutils';
import { HEATMAP_CHARTNAME } from '../core/chartname-constants';
import { SIZINGRANDOM_ANIMATION, FADERANDOM_ANIMATION, FADE_ANIMATION, SIZING_ANIMATION } from '../core/animation-constants';
import { getFillColor, getStrokeColor } from '../utils/colorutils';
import { getSVGsymbol } from '../utils/svg-symbols';
import { appendEle, roundedRect } from '../utils/zcsvgutils';
import { PATH_ELEMENT } from '../core/element-constants';
import { getHeatCellPosition } from '../events/events-utils';
import { d3_values, utils } from '../utils/zcutils';
import { d3_merge, d3_select } from '../core/d3-constants';
import { mathRandom, mathMin } from '../core/zc-constants';
import { colorscale_helpers } from '../axis/colorscale-helpers';
import { labelRenderer } from '../core/main-zc-init';
import { BarHighlighter } from '../bar/bar-highlight';
import { HeatmapHighlighter } from './heatmap-highlight';

/**
 *
 * @returns {chart} heatmap renderer.
 */

//hierarchy series name on the top like fusion chart
svg_plot_renderer.heatmap = function() {
    var isredraw = false,
        commonRendererProp;

    function chart(selection) {
        selection.each(function(data) {
            //renderer config values
            var rp = commonRendererProp.params;
            var chartObj = rp.chartObj,
                xscale = rp.xscale,
                yscale = rp.yscale,
                index = rp.datasetIndex,
                commonPlotOptions = rp.commonPlotOptions,
                labelQueue = rp.labelQueue,
                yaxiscolumnorder = rp.yaxiscolumnorder;

            var currentdata = data[index];
            var currentseriesdata = currentdata.data;

            //plotoptions

            var plotoptions = currentdata.plotoptions,
                gvp = (rp.gvp = utils.getValueOfPlotOptions(plotoptions, commonPlotOptions));

            //Animation variables
            var animationOptions = getAnimationConf(commonPlotOptions, null, chartObj, HEATMAP_CHARTNAME);
            var isAnimationEnable = animationOptions.enabled,
                animationType = animationOptions.type,
                easingType = animationOptions.easingType,
                transitionDuration = isredraw ? (isAnimationEnable ? 200 : 0) : animationOptions.duration,
                isRandomAnimation = animationType === SIZINGRANDOM_ANIMATION || animationType === FADERANDOM_ANIMATION,
                delay = isredraw
                    ? 0
                    : isRandomAnimation
                    ? function() {
                          return mathRandom() * ((transitionDuration * 3) / 5);
                      }
                    : 0,
                duration = isredraw ? transitionDuration : isRandomAnimation ? (transitionDuration * 2) / 5 : transitionDuration;

            var fillColor = (chart.fillColorFunc = getFillColor(chartObj, index)),
                strokeColor = getStrokeColor(chartObj, index),
                borderRadius = chartObj.dataObject.getTransRectValue('radius', null, null, null, null, commonPlotOptions.borderRadius),
                shape = gvp('symbol'),
                symbolPath = shape && getSVGsymbol(shape),
                xCategories = xscale.domain(),
                yCategories = yscale.domain(),
                hasColorScale = colorscale_helpers.isQuantileColorScale(chartObj);

            var container = d3_select(this);
            var series = appendEle(container, 'g', [1], '.', 'class', `heatmapseries_${index}`);
            chart.seriesdataAfterUpdate = [];

            currentseriesdata.forEach(function(d, i) {
                var serData = (chart.seriesdataAfterUpdate[i] = chartObj.dataObject.updateSeriesdatabyCategories(
                    d3_values(d),
                    xCategories,
                    yCategories,
                    yaxiscolumnorder
                ));
                var subseries = appendEle(series, 'g', !hasColorScale && currentdata.disabled ? [] : [1], '.', 'class', `subseries_${i}`);
                var path = appendEle(subseries, PATH_ELEMENT, serData);

                path.styles({
                    fill: function(d, i) {
                        var pos = getHeatCellPosition(chartObj, d, xscale, yscale, yaxiscolumnorder);
                        return fillColor(index, d, null, null, null, [pos.width, pos.height]);
                    },
                    'fill-opacity': commonPlotOptions.fillOpacity,
                    stroke: function(d) {
                        return strokeColor(index, d, commonPlotOptions.strokeColor);
                    },
                    'stroke-width': commonPlotOptions.strokeWidth,
                    'stroke-opacity': commonPlotOptions.strokeOpacity
                })
                    .attr('transform', function(d) {
                        var pos = getHeatCellPosition(chartObj, d, xscale, yscale, yaxiscolumnorder);
                        return `translate(${pos.x + pos.width / 2},${pos.y + pos.height / 2})`;
                    })
                    .attr('d', function(d) {
                        var pos = getHeatCellPosition(chartObj, d, xscale, yscale, yaxiscolumnorder);
                        if (symbolPath) {
                            return symbolPath(mathMin(pos.width, pos.height));
                        } else {
                            return roundedRect(-pos.width / 2, -pos.height / 2, pos.width, pos.height, borderRadius);
                        }
                    });

                //animation
                if (!isredraw) {
                    cancelAnimation(path); //REFER: heatmap redraw- animation issue
                    path.attr('opacity', 0) //#ZC892
                        .transition()
                        .delay(delay)
                        .duration(duration)
                        .ease(easingType)
                        .on('start', function() {
                            d3_select(this).attr('opacity', null); //#ZC892
                        })
                        .call(isRandomAnimation ? fadeTween : function() {})
                        .call(
                            animationType === SIZING_ANIMATION || animationType === SIZINGRANDOM_ANIMATION
                                ? scaleTween
                                : animationType === FADE_ANIMATION || animationType === FADERANDOM_ANIMATION
                                ? fadeTween
                                : flipTween
                        );
                }
            });

            if (labelRenderer) {
                labelQueue.push({
                    renderer: chart,
                    charttype: 'hierarchy',
                    chartObj: chartObj,
                    args: [currentdata, index, null, null, null, d3_merge(chart.seriesdataAfterUpdate)]
                });
            }

            chartObj.timerObj[chartObj.timerObj.length] = chart.renderedTimer = setTimeout(function() {
                isredraw = true;
                chart.chartRendered = true;
            }, transitionDuration + 17);
        });
        return chart;
    }
    chart.getHighlighter = function(chartObj, index) {
        if (!chart.highlighter) {
            chart.highlighter = new HeatmapHighlighter(chartObj, chart);
        }
        return chart.highlighter;
    };
    chart.commonRendererProp = function(_) {
        if (!arguments.length) {
            return commonRendererProp;
        }
        commonRendererProp = _;
        return chart;
    };
    return chart;
};

export var _heatmapChart = svg_plot_renderer.heatmap;
