// @zohocharts/zc/src/behaviours/visual-effects/brush-handles/brush-handle-options.js
//$Id$
//slider-marker-options.js

export function getDefaultSliderOptions(dimension, isHorizontal, segmentCount) {
    return {
        line: {
            strokeColor: 'grey',
            strokeWidth: 2,
            height: dimension[isHorizontal ? 1 : 0] + 20
        },
        triangleUp: {
            size: 15,
            fillColor: '#4C4C4C'
        },
        triangleDown: {
            size: 15,
            fillColor: '#4C4C4C'
        },
        triangleUpDown: {
            size: 15,
            fillColor: '#4C4C4C',
            padding: dimension[isHorizontal ? 1 : 0] + 15
        },
        box: {
            strokeWidth: 1,
            strokeColor: 'black',
            fillColor: 'transparent',
            height: dimension[isHorizontal ? 1 : 0],
            width: segmentCount ? dimension[isHorizontal ? 0 : 1] / segmentCount : 10
        },
        enlarge: {
            range: 1.2,
            height: dimension[isHorizontal ? 1 : 0],
            width: segmentCount ? dimension[isHorizontal ? 0 : 1] / segmentCount : 10
        },
        lever: {
            height: 26,
            width: 9,
            radius: 2,
            stripeHeight: 8,
            stripePadding: 2,
            fillColor: 'white',
            strokeColor: '#DADADA',
            stripeColor: '#DADADA',
            shadow: '0 0 2px rgba(0,0,0,0.2)'
        },
        hollowCircle: {
            size: 25,
            outerFillColor: 'white',
            outerStrokeColor: '#DADADA',
            innerStrokeColor: '#DADADA',
            shadow: '0 0 2px rgba(0,0,0,0.2)'
        },
        circle: {
            size: 25,
            outerStrokeColor: 'white',
            outerStrokeWidth: 2,
            innerFillColor: 'white',
            shadow: '0 0 2px rgba(0,0,0,0.2)'
        },
        vernier: {
            height: 26,
            width: 9,
            fillColor: 'white',
            strokeColor: '#DADADA',
            shadow: '0 0 2px rgba(0,0,0,0.2)'
        },
        pentagon: {
            size: 8,
            bandWidth: 3,
            fillColor: 'white',
            strokeColor: '#DADADA',
            shadow: '0 0 2px rgba(0,0,0,0.2)',
            height: dimension[isHorizontal ? 1 : 0]
        },
        lineCircle: {
            size: 15,
            strokeWidth: 2,
            strokeColor: 'blue',
            position: isHorizontal ? 'end' : 'start'
        },
        rectangle: {
            fillColor: '#333',
            strokeWidth: 1
        }
    };
}
