// @zohocharts/zc/src/behaviours/visual-effects/brush-handles/handles.js
import { getBorderRadius, splat } from '../../../utils/zcutils';
import { roundedRect } from '../../../utils/zcsvgutils';
import { getSVGsymbol } from '../../../utils/svg-symbols';
import { getDefaultSliderOptions } from './brush-handle-options';
import { d3_map } from '../../../core/d3-constants';

const __CUSTOM_SLIDER_SHAPES = d3_map();

/**
 *
 * @param name
 * @param handle
 */
export function setCustomBrushHandle(name, handle) {
    __CUSTOM_SLIDER_SHAPES.set(name, handle);
}

export const getSliderShape = (type, config, brushType, dimension, offset = [0, 0]) => {
    // user defined handles
    if (__CUSTOM_SLIDER_SHAPES.has(type)) {
        return __CUSTOM_SLIDER_SHAPES.get(type);
    }

    let handle = null,
        transform,
        index = { x: 1, y: 0 }[brushType],
        isHorizontal = brushType === 'x',
        translate = getSliderPosition(brushType, type, dimension, config.size);

    // add the offset
    // #ZC2227
    translate = translate.map((t0, i) => t0 + offset[i]);

    switch (type) {
        case 'line':
            handle = {
                shape: getSVGsymbol('vertical-line')(config.height),
                style: {
                    stroke: config.strokeColor,
                    'stroke-width': config.strokeWidth
                }
            };
            break;
        case 'triangleUp':
            handle = {
                shape: getSVGsymbol('triangle-down')(config.size),
                style: {
                    fill: config.fillColor,
                    stroke: config.strokeColor,
                    'stroke-width': config.strokeWidth
                }
            };
            break;

        case 'triangleDown':
            handle = {
                shape: getSVGsymbol('triangle-up')(config.size),
                style: {
                    fill: config.fillColor,
                    stroke: config.strokeColor,
                    'stroke-width': config.strokeWidth
                }
            };
            break;
        case 'triangleUpDown':
            var padding = config.padding / 2,
                transformDown = isHorizontal ? [0, padding] : [padding, 0],
                transformUp = transformDown.map(t0 => t0 * -1);

            transformUp = [transformUp[0] + translate[0], transformUp[1] + translate[1]];
            transformDown = [transformDown[0] + translate[0], transformDown[1] + translate[1]];

            handle = {
                shape: [getSVGsymbol('triangle-down')(config.size), getSVGsymbol('triangle-up')(config.size)],
                style: [
                    {
                        fill: config.fillColor,
                        stroke: config.strokeColor,
                        'stroke-width': config.strokeWidth,
                        transform: `translate(${transformUp})`
                    },
                    {
                        fill: config.fillColor,
                        stroke: config.strokeColor,
                        'stroke-width': config.strokeWidth,
                        transform: `translate(${transformDown})`
                    }
                ]
            };
            break;
        case 'box':
            var strokeWidth = config.strokeWidth,
                w = config.width,
                h = config.height;

            handle = {
                shape: roundedRect(-(w - strokeWidth) / 2, -(h - strokeWidth) / 2, w - strokeWidth, h - strokeWidth, getBorderRadius()),
                style: {
                    stroke: config.strokeColor,
                    'stroke-width': config.strokeWidth,
                    fill: config.fillColor
                }
            };
            break;
        case 'enlarge':
            var strokeWidth = config.strokeWidth,
                range = config.range || 1,
                w = config.width * range,
                h = config.height * range;

            handle = {
                shape: roundedRect(-w / 2, -h / 2, w, h, getBorderRadius()),
                style: {
                    stroke: config.strokeColor,
                    'stroke-width': config.strokeWidth,
                    fillColor: config.fillColor
                }
            };
            break;
        case 'lever':
            var r = getBorderRadius(config.radius);
            handle = {
                shape: [
                    roundedRect(-config.width / 2, -config.height / 2, config.width, config.height, r),
                    `M${-config.stripePadding / 2} ${-config.stripeHeight / 2} L${-config.stripePadding / 2} ${config.stripeHeight /
                        2} M${config.stripePadding / 2} ${-config.stripeHeight / 2} L${config.stripePadding / 2} ${config.stripeHeight / 2}`
                ],
                style: [
                    {
                        fill: config.fillColor,
                        stroke: config.strokeColor
                    },
                    {
                        stroke: config.stripeColor
                    }
                ]
            };
            break;
        case 'hollowCircle':
            handle = {
                shape: [getSVGsymbol()(config.size), getSVGsymbol()(config.size * 0.4)],
                style: [
                    {
                        fill: config.outerFillColor,
                        stroke: config.outerStrokeColor
                    },
                    {
                        fill: config.innerFillColor,
                        stroke: config.innerStrokeColor
                    }
                ]
            };
            break;
        case 'circle':
            handle = {
                shape: [getSVGsymbol()(config.size), getSVGsymbol()(config.size * 0.3)],
                style: [
                    {
                        fill: config.outerFillColor,
                        stroke: config.outerStrokeColor,
                        'stroke-width': config.outerStrokeWidth
                    },
                    {
                        fill: config.innerFillColor,
                        stroke: config.innerStrokeColor
                    }
                ]
            };
            break;
        case 'vernier':
            var w = config.width,
                h = config.height,
                path1 = `M${-w * 0.5} ${-h * 0.5} L${w * 0.5} ${-h * 0.5} L${w * 0.5} ${h * 0.5} L${-w * 0.5} ${h * 0.2} Z`,
                path2 = `M${-w * 0.5} ${-h * 0.5} L${w * 0.5} ${-h * 0.5} L${w * 0.5} ${h * 0.2} L${-w * 0.5} ${h * 0.5} Z`;

            handle = {
                shape: [[path1], [path2]],
                style: {
                    fill: config.fillColor,
                    stroke: config.strokeColor
                }
            };
            break;
        case 'pentagon':
            var size = config.size,
                bandWidth = config.bandWidth,
                h = config.height,
                path =
                    `M${-size * 0.5 - bandWidth} ${-h * 0.5 - bandWidth} L${size * 0.5 + bandWidth} ${-h * 0.5 - bandWidth} L${size * 0.5 +
                        bandWidth} ${h * 0.5}L0 ${h * 0.5 + bandWidth * 1.5 + 2} L${-size * 0.5 - bandWidth} ${h * 0.5}Z` +
                    `M${-size * 0.5} ${-h * 0.5} L${size * 0.5} ${-h * 0.5} L${size * 0.5} ${h * 0.5 - bandWidth * 0.5}L0 ${h * 0.5 +
                        bandWidth * 0.5} L${-size * 0.5} ${h * 0.5 - bandWidth * 0.5}Z`,
                innerBand = `M${-size * 0.5} ${-h * 0.5} L${size * 0.5} ${-h * 0.5} L${size * 0.5} ${h * 0.5} L${-size * 0.5} ${h * 0.5}Z`;

            handle = {
                _effectElementIndex: 1,
                shape: [innerBand, path],
                style: [
                    {},
                    {
                        'fill-rule': 'evenodd',
                        fill: config.fillColor,
                        stroke: config.strokeColor
                    }
                ]
            };
            break;
        case 'lineCircle':
            var size = config.size,
                lineSize = dimension[index],
                factor = config.position === 'end' ? 1 : -1,
                circleTranslate = (size / 2) * factor,
                lineType = 'vertical-line';

            // add the size if the marker needs in the end
            if (config.position === 'end') {
                circleTranslate += dimension[index];
            }

            transform = [0, circleTranslate];

            if (!isHorizontal) {
                transform.reverse();
            }
            // #ZC2227
            transform = transform.map((t0, i) => t0 + offset[i]);

            handle = {
                shape: [getSVGsymbol('circle')(size), getSVGsymbol(lineType)(lineSize)],
                style: [
                    {
                        fill: config.fillColor,
                        stroke: config.strokeColor,
                        transform: `translate(${transform.join(' ')})`
                    },
                    {
                        fill: config.fillColor,
                        stroke: config.strokeColor,
                        'stroke-width': config.strokeWidth
                    }
                ]
            };
            break;

        case 'rectangle': {
            //  use default
            // if rectangle, use the default styles
            return {
                shape: 'default',
                style: { ...config }
            };
        }
    }

    transform = (translate && `translate(${translate.join(' ')}) `) || '';
    splat(handle.style).forEach(style => {
        let currentTransform = style.transform || transform;
        if (brushType === 'y') {
            currentTransform += ' rotate(-90)';
        }
        style.transform = currentTransform;
    });

    return handle;
};

export const getSliderPosition = (brushType, sliderType, dimension, size) => {
    let index = { x: 1, y: 0 }[brushType],
        transform = [0, 0];

    switch (sliderType) {
        case 'triangleUp':
            transform[index] = -size / 2;
            break;
        case 'triangleDown':
            transform[index] = dimension[index] + size / 2;
            break;
        default:
            transform[index] = dimension[index] / 2;
    }
    return transform;
};

/**
 *
 * @param brushType
 * @return {*}
 */
export function getD3BrushHandle(brushType) {
    return { x: [{ type: 'w' }, { type: 'e' }], y: [{ type: 'n' }, { type: 's' }] }[brushType];
}

/**
 *
 * @param brushType
 * @param sliderType
 * @param extent
 * @param segmentCount
 * @param userStyles
 */
export function getMergedHandleStyles(brushType, sliderType, extent, segmentCount, userStyles) {
    let dimension = [extent[1][0] - extent[0][0], extent[1][1] - extent[0][1]],
        defaultStyles = getDefaultSliderOptions(dimension, brushType === 'x', segmentCount)[sliderType];

    return { ...defaultStyles, ...userStyles };
}

/**
 *
 * @param brushType
 * @param sliderType
 * @param extent
 * @param segmentCount
 * @param userStyles
 * @return {{shape: string, style: {'stroke-width': *, stroke: *}}|{shape: string, style: {transform: string, 'stroke-width': *, fill: *, stroke: *}}|{shape: [string, string], style: [{transform: string, 'stroke-width': *, fill: *, stroke: *}, {transform: string, 'stroke-width': *, fill: *, stroke: *}]}|{shape: string, style: {'stroke-width', fill, stroke: *}}|{shape, style: {fillColor, 'stroke-width', stroke}}|*}
 */
function getBrushHandleStyles(brushType, sliderType, extent, segmentCount, userStyles, offset) {
    let dimension = [extent[1][0] - extent[0][0], extent[1][1] - extent[0][1]],
        defaultStyles = getDefaultSliderOptions(dimension, brushType === 'x', segmentCount)[sliderType],
        handleStyles = { ...defaultStyles, ...userStyles },
        handles = getSliderShape(sliderType, handleStyles, brushType, dimension, offset);

    splat(handles).forEach(handle => {
        // splat(handle.style).forEach(style => {
        //     // style.transform = style.transform || null;
        // });
        handle.alignCenter = false;
    });

    return handles;
}

/**
 *
 * @param brushType
 * @param sliderType
 * @param extent
 * @param segmentCount
 * @param styles
 * @return {*}
 */
export function getBrushHandle(brushType, sliderType = 'lever', extent, segmentCount = null, styles, offset) {
    let handleShapes = getBrushHandleStyles(brushType, sliderType, extent, segmentCount, styles, offset);
    return getD3BrushHandle(brushType).map(function(handle) {
        return { ...handle, ...handleShapes };
    });
}
