// @zohocharts/zc/src/mark/helper/custom-mark/div-mark.js
//
import { defined, pick } from '../../../utils/zcutils';
import { getHTML4SVGSymbol, getSVGSymbolString } from '../../../utils/svg-symbols';
import { getMarkValue } from '../../parser/parser-helpers';

function divMark(config, cache) {
    let { shape = 'square' } = config,
        useSVGImage = shape !== 'square' && shape !== 'rect';
    return {
        config: {
            background: function(d, i){
                let node = this,
                    args = [d, i];

                if(defined(config.background)){
                    return getMarkValue(node, args, config, 'background', cache);
                }
                if(useSVGImage){
                    let _width = getMarkValue(node, args, config, 'width', cache),
                        _height = getMarkValue(node, args, config, 'height', cache),
                        _strokeColor = getMarkValue(node, args, config, 'strokeColor', cache),
                        _fillColor = getMarkValue(node, args, config, 'fillColor', cache),
                        _shape = getMarkValue(node, args, config, 'shape', cache);

                    return getSVGSymbolString(
                        getHTML4SVGSymbol({ ...config, width: _width, height: _height, strokeColor: _strokeColor, fillColor: _fillColor,shape:  _shape}, true),
                        _width,
                        _height
                    );
                }else {
                    let _fillColor = getMarkValue(node, args, config, 'fillColor', cache);
                    return _fillColor;
                }
            },
            backgroundSize: function(d, i){
                let node = this,
                    args = [d, i];

                if(defined(config.backgroundSize)){
                    return getMarkValue(node, args, config, 'backgroundSize', cache);
                }
            },
            borderRadius: function(d, i){
                let node = this,
                    args = [d, i];

                if(defined(config.borderRadius)){
                    return `${getMarkValue(node, args, config, ['radius', 'borderRadius'], cache)}px`;
                }

                if(useSVGImage){
                    return null;
                }else{
                    return `${getMarkValue(node, args, config, ['radius', 'borderRadius'], cache)}px`;
                }
            },
            border: function(d, i){
                let node = this,
                    args = [d, i];

                if(defined(config.border)){
                    return getMarkValue(node, args, config, 'border', cache);
                }
                if(useSVGImage){
                    return 'none';
                }else{
                    let _strokeWidth = getMarkValue(node, args, config, 'strokeWidth', cache),
                        _strokeColor = getMarkValue(node, args, config, 'strokeColor', cache);
                    return `${pick(_strokeWidth, 1)}px solid ${_strokeColor}`; // TODO: dashStyle can be applied
                }
            },
            width: function(d, i) {
                let node = this,
                    args = [d, i],
                    _width = getMarkValue(node, args, config, 'width', cache);

                return `${_width}px`;
            },
            height: function(d, i) {
                let node = this,
                    args = [d, i],
                    _height = getMarkValue(node, args, config, 'height', cache);
                return `${_height}px`;
            }
        },
        depends: {
            background: useSVGImage ? ['width', 'height', 'strokeColor', 'fillColor', 'shape'] : ['fillColor'],
            backgroundSize: [],
            borderRadius: useSVGImage ? [] : ['radius', 'borderRadius'],
            border: useSVGImage ? [] : ['strokeWidth', 'strokeColor'],
            width: ['width'],
            height: ['height']
        }
    };
}

export { divMark };
