// @zohocharts/zc/src/scale-parser/legend/shape/shape-domain-parser.js
//$Id$
import { NUMERIC_DATATYPE, TIME_DATATYPE } from '../../../core/datatype-constants';
import { UNDEFINED } from '../../../core/zc-constants';
import { d3_merge, d3_ascending } from '../../../core/d3-constants';
import { d3_values } from '../../../utils/zcutils';
import { ScaleRegistry } from '../../../scale/scale-registry';
import { DataProcessor } from '../../../core/data-processor';
import { fetchForShapeDomainParser } from '../../fetch-operators';
import { convertNull } from '../../../core/data-processor/size/convert-null';
import { categories } from '../../../scale/domain/categories';
import { toTimestamp } from '../../../scale/scale-component';

export function shapeDomainParser(encodeType, index, config, chart) {
    let processor = chart.dataObject,
        dataSet = chart.dataset,
        dataType = dataSet.getDataType(encodeType, index),
        columnInfo = dataSet.getColumn(encodeType, index),
        nullAsOptions = { config, columnKey: dataSet.getColumnIdx(encodeType, index) },
        _data = processor.processedseriesData,
        //
        isCatNull = DataProcessor.helpers.isCatNull,
        isCatNullX = isCatNull(dataSet, 'x'),
        isCatNullY = isCatNull(dataSet, 'y', 0),
        condition =
            hasGeo(chart) || (chart.systemConf.legend.singleLayout && dataType === NUMERIC_DATATYPE)
                ? false
                : d => !(isCatNullX(dataSet.getActualX(d, index)) || isCatNullY(dataSet.getY(d, index))),
        schema = dataType === TIME_DATATYPE ? toTimestamp(config, columnInfo) : null,
        { minRange, maxRange, stops, ranges } = schema || config,
        nullAsConvertion = d => convertNull(d[columnInfo.dataindex], nullAsOptions);

    let domainOperators = {},
        defaultOperators = fetchForShapeDomainParser(dataType);

    /*
        EXTREMES
     */
    if (defaultOperators.has('extremes')) {
        if (chart.systemConf.legend.singleLayout && encodeType === 'z') {
            domainOperators.extremes = [() => processor.getMinMaxZ()];
        } else {
            domainOperators.extremes = [
                {
                    _type: 'extremesByAccessor',
                    data: _data,
                    condition,
                    accessor: function() {
                        return nullAsConvertion;
                    }
                }
            ];
        }
    }

    if (defaultOperators.has('categories')) {
        let _d = d3_merge(_data.map(_ => d3_merge(_.data)));
        domainOperators.categories = categories(_d, nullAsConvertion);
    }

    if (dataType === NUMERIC_DATATYPE && !chart.isGeoMap) {
        let base = config.base;
        domainOperators.base = base !== UNDEFINED ? base : 0;
    }

    /*
        Custom min max range
     */
    domainOperators.min = minRange;
    domainOperators.max = maxRange;

    if (ranges && ranges.length) {
        /* Ranges*/
        domainOperators.sortedRanges = {
            datatype: dataType,
            clone: false,
            config: { ...config, ranges }
        };
    } else if (stops && stops.length) {
        /*Stops*/
        domainOperators.filteredStops = {
            array: d3_values(stops).sort(d3_ascending),
            min: minRange,
            max: maxRange
        };
    }
    return domainOperators;
}

ScaleRegistry.add('shape.domainParser', shapeDomainParser);

function hasGeo(chartObj) {
    // TODO  change chartObj.dataObject.isFlowChartCategory
    return chartObj.isGeoMap || chartObj.mapObj || chartObj.dataObject.isFlowChartCategory;
}
