// @zohocharts/zc/src/scale-parser/legend/shape/shape-legend-parser.js
//$Id$
import { Registry } from '../../../core/zc-registry';
import { ScaleRegistry } from '../../../scale/scale-registry';

export function shapeLegendParser(encodeType, index = 0, chart, config) {
    let dataSet = chart.dataset,
        dataType = dataSet.getDataType(encodeType, index),
        columnInfo = dataSet.getColumn(encodeType, index);

    var parser = {
        type: dataType,
        dataType: dataType,
        scaleType: columnInfo.scaleType,
        field: columnInfo,
        reverse: config.reversed,
        clamp: true,
        data: [],
        domain: {
            operators: ScaleRegistry.get('shape.domainParser')(encodeType, index, config, chart),
            evaluator: ScaleRegistry.get('shape.evalDomain')
        },
        range: {
            operators: ScaleRegistry.get('shape.rangeParser')(encodeType, index, config, chart),
            evaluator: ScaleRegistry.get('shape.evalRange')
        }
    };

    return parser;
}

Registry.setComponent('shape.parser', shapeLegendParser);
