// @zohocharts/zc/src/legend/events/visual-effects/discrete/update-plot-by-selection.js
import { d3_event } from '../../../../core/d3-constants';
import { docTapFunction } from '../../../../events/document-events';

export function filterPlotOnClick(mouse, data) {
    var legendComponent = this.options.legendInstance.component.discrete_encode,
        point = data.point,
        legendOptions = data.eventManager.options.config;

    var isTouchEvent = d3_event() && d3_event().touches;
    if (isTouchEvent) {
        docTapFunction();
    }
    var race = this.chart.racing;
    if(race && !race.ended && race.started){
        race.legendFilterInfo = {
            callback: legendComponent.toggleItemNUpdateChart,
            args: [point.ele, point.data, point.i, legendOptions],
            isLegendFiltered: true
        }
    } else {
        if(race){
            this.chart.racing.syncTransition = this.chart.container.transition().duration(0);
        }
        legendComponent.toggleItemNUpdateChart(point.ele, point.data, point.i, legendOptions);
    }
}

export function highlightSeries(mouse, data) {
    var legendComponent = this.options.legendInstance.component.discrete_encode,
        point = data.point,
        legendOptions = data.eventManager.options.config;

    legendComponent.highlightSeries(point.ele, point.data, point.i, legendOptions, data.chart, true);
    legendComponent.highlightItem(point.ele, point.data, point.i, legendOptions, true);
}
export function resetHighlightSeries(mouse, data) {
    var legendComponent = this.options.legendInstance.component.discrete_encode,
        point = data.point,
        legendOptions = data.eventManager.options.config;

    legendComponent.highlightSeries(point.ele, point.data, point.i, legendOptions, data.chart);
    legendComponent.highlightItem(point.ele, point.data, point.i, legendOptions);
}
