// @zohocharts/zc/src/wordcloud/canvasWordcloudchart.js
//$Id$
//canvasWordcloudchart.js
import { Registry } from '../core/zc-registry';
import { wordcloudLayout } from './wordcloud-layout';
import { getFillColor, getStrokeColor, isMultiColoring, isWithinColorRange, getRawColor } from '../utils/colorutils';
import { isObjectPropDefined, pick, defined } from '../utils/zcutils';
import { NULL, PX, UNDEFINED } from '../core/zc-constants';
import { CENTER_ALIGN } from '../core/align-constants';
import { effects } from '../effects/zceffects';
import { wordCloudUtils } from './wordcloud-utils';
import { mathMin } from '../core/zc-constants';
import { colorscale_helpers } from '../axis/colorscale-helpers';
// import { legend_helpers } from '../legend/legend-helper';
import { legend_helpers, canvas_plot_renderer } from '../core/main-zc-init';
import { canvasUtils } from '../utils/canvasutils';
import { getUserFontValue } from '../label/font-helpers';
import { _canvasDatalabelUtils } from '../label/canvas-datalabel-utils';
import { _wordcloudBoundBoxCanvas } from './wordcloud-boundbox-canvas';
import { drawMaskBackgroundInCanvas } from './wordcloud-mask-background-canvas';
import { isFacet } from '../axis/hierarchy-axis-utils';
import { WordcloudHighlighter } from './wordcloud-highlight';

let annotation = Registry.getComponent('annotation');
canvas_plot_renderer.wordcloud = function() {
    var _getSeriesIndex = wordCloudUtils.getSeriesIndex,
        _getItemIndex = wordCloudUtils.getItemIndex,
        helpers = _canvasDatalabelUtils,
        commonRendererProp;

    function chart(context, seriesdata, effect, filterIndex, encode, _index) {
        var rp = commonRendererProp.params,
            chartObj = rp.chartObj,
            commonPlotOptions = rp.commonPlotOptions,
            notesArray = rp.notesArray,
            renderer = chartObj.renderer[rp.rendererIndex || 0],
            layoutUpate = chartObj.phase !== 'legendFilter-colorScale' && !effect,
            layoutObj = layoutUpate ? new wordcloudLayout(commonRendererProp, seriesdata) : renderer.layout,
            dimension = layoutObj.dimension,
            w = dimension.width,
            h = dimension.height,
            processedData = layoutObj.processedData,
            fillColor = [],
            strokeColor = [],
            hasColorScale = colorscale_helpers.isQuantileColorScale(chartObj),
            multiColoring = isMultiColoring(commonPlotOptions, chartObj),
            highlightStrokeWidth,
            highlightStrokeColor,
            highlightFillColor,
            highlightIndex,
            font = [],
            textShadow = [],
            paneid = isFacet(chartObj) && `${seriesdata[0].row}_${seriesdata[0].column}`;

        if (effect) {
            highlightIndex = _index; //isObjectPropDefined(seriesdata, 'data[0][0].seriesIndex', true); //ZC1440
            highlightStrokeWidth = effect['stroke-width'];
            highlightStrokeColor = effect.stroke;
            highlightFillColor = effect.fill;
        }

        chart.total = layoutObj.sum;
        chart.extentY = layoutObj.fontScale.domain();

        layoutObj.data.forEach((d, i) => {
            fillColor[i] = chart.fillColorFunc = getFillColor(chartObj, i, paneid, {
                x1: 0,
                x2: w,
                y1: 0,
                y2: h,
                r: mathMin(w, h)
            });
            strokeColor[i] = getStrokeColor(chartObj, i, paneid);
        });

        if (layoutUpate) {
            layoutObj.drawLayout(draw);
            renderer.layout = layoutObj;
        } else {
            draw();
        }
        chart.layout = layoutObj;
        layoutObj.paneid = isFacet(chartObj) ? `${seriesdata[0].row}_${seriesdata[0].column}` : 0;

        function draw() {
            var transformation = wordCloudUtils.getContainerTransformation(layoutObj, dimension);
            let scaleFactor = layoutObj.output.scaleFactor,
                fontScaleFactor = layoutObj.isScalable ? 1 : scaleFactor;

            context.save();
            context.translate(transformation.translate[0], transformation.translate[1]);
            context.scale(transformation.scale, transformation.scale);
            drawMaskBackgroundInCanvas(context, commonPlotOptions, layoutObj, layoutUpate);

            if (defined(highlightIndex)) {
                processedData = processedData.filter(function(d) {
                    return _getSeriesIndex(d) === highlightIndex;
                });

                if (hasColorScale) {
                    processedData = legend_helpers.filterSelectionByColor(processedData, true, filterIndex, chartObj);
                }
            }

            processedData.forEach(function(d, i) {
                var layout = d.layout;
                if (!layout) {
                    return;
                }

                var text = layout.text,
                    x = layout.x * fontScaleFactor,
                    y = layout.y * fontScaleFactor,
                    dy = wordCloudUtils.getDy(layout),
                    seriesIndex = _getSeriesIndex(d),
                    itemIndex = _getItemIndex(d),
                    datalabelOptions = layoutObj.datalabelOptions[seriesIndex],
                    plotOptions = (effect ? seriesdata : seriesdata[seriesIndex]).plotoptions || {},
                    tX,
                    tY;

                font[seriesIndex] = font[seriesIndex] || canvasUtils.getFontStyles(chartObj, datalabelOptions);
                textShadow[seriesIndex] = textShadow[seriesIndex] || helpers.parseTextShadow(datalabelOptions.textShadow);
                var strokeWidth = pick(highlightStrokeWidth, plotOptions.strokeWidth, commonPlotOptions.strokeWidth, 1),
                    stroke = pick(
                        highlightStrokeColor,
                        plotOptions.strokeColor !== UNDEFINED ? plotOptions.strokeColor : commonPlotOptions.strokeColor
                    ),
                    strokeOpacity = pick(plotOptions.strokeOpacity, commonPlotOptions.strokeOpacity),
                    fillOpacity = pick(plotOptions.fillOpacity, commonPlotOptions.fillOpacity),
                    gradientOptions = fillColor[seriesIndex].gradientOptions;

                var drawText = function(context, fillStyle) {
                    font[seriesIndex]['font-size'] = layout.size * fontScaleFactor + PX;
                    context.save();

                    if (layout.rotation) {
                        context.translate(x, y);
                        context.rotate(layout.rotation);
                        context.translate(-x, -y);
                    }

                    if (!effect && wordCloudUtils.hasBoundBox(d)) {
                        _wordcloudBoundBoxCanvas(context, d, layoutObj, commonRendererProp);
                    }

                    helpers.addTextShadow(context, textShadow[seriesIndex]);
                    context.font = getUserFontValue(font[seriesIndex]);
                    context.textAlign = CENTER_ALIGN;

                    if (layout.nodeList.contentType === 'html') {
                        layout.nodeList.forEach(function(a) {
                            var userStyle = a.style,
                                content = a.text,
                                userFont,
                                fSize,
                                userFillStyle;

                            if (userStyle.length) {
                                userFont = getUserFontValue(font[seriesIndex], userStyle);
                                fSize = parseFloat(userStyle['font-size']);
                                userFillStyle = userStyle.fill;
                            }

                            tX = x + (a.dx || 0);
                            tY =
                                y +
                                (a.dy || 0) * layout.size * fontScaleFactor +
                                layout.size * fontScaleFactor * dy -
                                (wordCloudUtils.isMultiLine(layout) ? (layout.height * fontScaleFactor) / 2 : 0);

                            context.font = userFont || getUserFontValue(font[seriesIndex]);
                            context.strokeStyle = strokeColor[seriesIndex](
                                multiColoring ? itemIndex : seriesIndex,
                                d,
                                stroke,
                                NULL,
                                commonPlotOptions.strokeOpacity
                            );
                            context.lineWidth = strokeWidth;
                            context.strokeText(content, tX, tY);

                            context.fillStyle = userFillStyle || fillStyle;
                            context.fillText(content, tX, tY);
                        });
                    } else {
                        var content = layout.nodeList.innerSVG;
                        tX = x;
                        tY = y + layout.size * fontScaleFactor * dy;

                        context.strokeStyle = strokeColor[seriesIndex](
                            multiColoring ? itemIndex : seriesIndex,
                            d,
                            stroke,
                            NULL,
                            strokeOpacity
                        );
                        context.lineWidth = strokeWidth;
                        context.lineJoin = 'round';
                        context.strokeText(content, tX, tY);

                        context.fillStyle = fillStyle;
                        context.fillText(content, tX, tY);
                    }

                    context.restore();
                };

                var customGradientDimension = NULL;
                if (gradientOptions.type === 'linear' && gradientOptions.options[gradientOptions.type].gradientUnits !== 'userSpaceOnUse') {
                    customGradientDimension = {
                        x1: x - (layout.width * fontScaleFactor) / 2,
                        x2: x + (layout.width * fontScaleFactor) / 2,
                        y1: y - (layout.height * fontScaleFactor) / 2,
                        y2: y + (layout.height * fontScaleFactor) / 2
                    };
                }

                if (chartObj.imagePallete && isWithinColorRange(chartObj.dataset.getClr(d), chartObj)) {
                    effects.canvas_patterns.imageFill(
                        {
                            x: x,
                            y: y,
                            context: context,
                            name: getRawColor(chartObj, multiColoring ? itemIndex : seriesIndex, d, seriesIndex, paneid)
                        },
                        drawText
                    );
                } else {
                    var fillStyle =
                        highlightFillColor ||
                        fillColor[seriesIndex](
                            multiColoring ? itemIndex : seriesIndex,
                            d,
                            NULL,
                            NULL,
                            fillOpacity,
                            customGradientDimension
                        );
                    drawText(context, fillStyle);
                }
            });
            context.restore();
            //Annotation preprocessing

            if (!annotation._empty && !effect) {
                chartObj.seriesdata.forEach(function(series, i) {
                    notesArray = annotation.preprocessAnnotations(
                        chartObj.seriesdata,
                        i,
                        function(dt, k, j) {
                            var layout = dt.layout,
                                x = transformation.translate[0] + (layout.x - layout.width / 2) * scaleFactor,
                                y = transformation.translate[1] + (layout.y - layout.height * wordCloudUtils.getDy(layout)) * scaleFactor;

                            return [x, y];
                        },
                        notesArray,
                        chartObj,
                        rp
                    );
                });
            }

            chart.chartRendered = true;
        }
        return chart;
    }

    chart.commonRendererProp = function(_) {
        if (!arguments.length) {
            return commonRendererProp;
        }
        commonRendererProp = _;
        return chart;
    };
    chart.getHighlighter = function(chartObj, index) {
        return (chart.highlighter = new WordcloudHighlighter(chartObj, chart));
    };

    return chart;
};

export var _canvasWordcloudChart = canvas_plot_renderer.wordcloud;
