// @zohocharts/zc/src/wordcloud/font-scale-parser.js
//
import { ORDINAL_DATATYPE } from '../core/datatype-constants';
import { ScaleRegistry } from '../scale/scale-registry';
import { _legendType } from '../legend/legend-renderer-helpers';
import { fontRangeParser } from './font-range-parser';

function fontScaleParser(encodeType, index = 0, chart, config) {
    let dataSet = chart.dataset,
        dataType = dataSet.getDataType(encodeType, index),
        columnInfo = dataSet.getColumn(encodeType, index);

    var parser = {
        type: dataType,
        dataType: dataType,
        scaleType: dataType === ORDINAL_DATATYPE ? '_ordinal' : dataType,
        field: columnInfo,
        reverse: config.reversed,
        clamp: true,
        data: [],
        domain: {
            operators: ScaleRegistry.get('size.domainParser')(encodeType, index, config, chart),
            evaluator: ScaleRegistry.get('size.evalDomain')
        },
        range: {
            operators: fontRangeParser(encodeType, index, config, chart),
            evaluator: ScaleRegistry.get('size.evalRange')
        }
    };

    return parser;
}

export { fontScaleParser };
ScaleRegistry.add('wordcloud.font.parser', fontScaleParser);
