// @zohocharts/zc/src/wordcloud/wordcloudchart.js
//$Id$
//wordcloudchart.js
import { Registry } from '../core/zc-registry';
import { svg_plot_renderer } from '../core/main-zc-init';
import { wordCloudUtils } from './wordcloud-utils';
import { d3_select, d3_ease_linear } from '../core/d3-constants';
import { getAnimationConf } from '../utils/animationutils';
import { WORDCLOUD_CHARTNAME } from '../core/chartname-constants';
import { NULL, PX, UNDEFINED } from '../core/zc-constants';
import {
    FADEONEBYONE_ANIMATION,
    FADERANDOM_ANIMATION,
    FADEALL_ANIMATION,
    SIZING_ANIMATION,
    RANDOMOUT_ANIMATION
} from '../core/animation-constants';
import { WordcloudHighlighter } from './wordcloud-highlight';
import { appendEle } from '../utils/zcsvgutils';
import { applyDatalabelFonts, applyTextShadow4Firefox } from '../label/datalabelsUtils';
import { wordcloudLayout } from './wordcloud-layout';
import { mathMin, mathRandom } from '../core/zc-constants';
import { MIDDLE_ALIGN } from '../core/align-constants';
import { TEXT_ELEMENT, PATH_ELEMENT } from '../core/element-constants';
import { translateTspan } from '../label/tspan-utils';
import { radianToDegree } from '../polaraxis/polarutils';
import { pick, mergeJSON } from '../utils/zcutils';
import { getFillColor, getStrokeColor, isMultiColoring } from '../utils/colorutils';
import { _wordcloudBoundBoxSvg } from './wordcloud-boundbox-svg';
import { drawMaskBackgroundInSvg } from './wordcloud-mask-background-svg';
import { setOpacityForTickFilterElements, setPivotAttributes } from '../plot/svg/set-pivot-attr';
import { isFacet } from '../axis/hierarchy-axis-utils';

let annotation = Registry.getComponent('annotation');
/**
 *
 * @returns {chart} wordcloud renderer.
 */
// Neg value
// bound box
svg_plot_renderer.wordcloud = function() {
    var isredraw = false,
        commonRendererProp,
        layoutObj,
        _getSeriesIndex = wordCloudUtils.getSeriesIndex,
        _getItemIndex = wordCloudUtils.getItemIndex,
        filterAccessor = function(d) {
            d3_select(this).style('opacity', d.layout ? NULL : 0); //Set opacity:0 to the words that are not visible to improve redraw
            return d.layout;
        };

    function chart(selection) {
        selection.each(function(data) {
            var rp = commonRendererProp.params,
                chartObj = rp.chartObj,
                commonPlotOptions = rp.commonPlotOptions,
                notesArray = rp.notesArray,
                scaleFactor,
                layoutUpate = chartObj.phase !== 'legendFilter-colorScale',
                //Animation variables
                animationOptions = getAnimationConf(commonPlotOptions, NULL, chartObj, WORDCLOUD_CHARTNAME),
                isAnimationEnable = animationOptions.enabled,
                animationType = animationOptions.type,
                easingType = isredraw ? d3_ease_linear : animationOptions.easingType,
                transitionDuration = isredraw ? (isAnimationEnable ? 200 : 0) : animationOptions.duration,
                dataCount = chartObj.dataObject.totalnoofdatas,
                duration = isredraw
                    ? transitionDuration
                    : animationType === FADEONEBYONE_ANIMATION || animationType === FADERANDOM_ANIMATION
                    ? transitionDuration / dataCount
                    : transitionDuration,
                delay = isredraw
                    ? 0
                    : animationType === FADEONEBYONE_ANIMATION
                    ? function(d, i) {
                          return i * (transitionDuration / dataCount);
                      }
                    : animationType === FADERANDOM_ANIMATION
                    ? function() {
                          return mathRandom() * dataCount * (transitionDuration / dataCount);
                      }
                    : 0,
                initialOpacity =
                    animationType === FADEALL_ANIMATION ||
                    animationType === FADERANDOM_ANIMATION ||
                    animationType === FADEONEBYONE_ANIMATION
                        ? 0
                        : NULL,
                container = appendEle(d3_select(this), 'g', [data], '.', 'class', 'wordcloudgroup'),
                paneid = isFacet(chartObj) && `${data[0].row}_${data[0].column}`;

            chart.layout = layoutObj = layoutUpate ? new wordcloudLayout(commonRendererProp, data) : layoutObj; //TODO: on legend brushing need not to create new layout
            var dimension = layoutObj.dimension,
                w = dimension.width,
                h = dimension.height,
                fillColor = [],
                strokeColor = [],
                multiColoring = isMultiColoring(commonPlotOptions, chartObj);

            data.forEach((d, i) => {
                fillColor[i] = chart.fillColorFunc = getFillColor(chartObj, i, paneid, { x1: 0, x2: w, y1: 0, y2: h, r: mathMin(w, h) });
                strokeColor[i] = getStrokeColor(chartObj, i, paneid);
            });

            chart.total = layoutObj.sum;
            chart.extentY = layoutObj.fontScale.domain(); //percent, maxpercent only. No splitpercent

            if (layoutUpate) {
                layoutObj.drawLayout(draw);
            } else {
                draw();
            }

            function draw() {
                drawMaskBackgroundInSvg(container, commonPlotOptions, layoutObj);
                scaleFactor = layoutObj.output.scaleFactor;
                var transformation = wordCloudUtils.getContainerTransformation(layoutObj, dimension),
                    fontScaleFactor = layoutObj.isScalable ? 1 : scaleFactor;

                data.forEach((d, i) => {
                    let plotOptions = chartObj.seriesdata[i].plotoptions;
                    let gradientOptions =
                        plotOptions && plotOptions.gradients
                            ? mergeJSON(commonPlotOptions.gradients, plotOptions.gradients, true)
                            : commonPlotOptions.gradients;

                    fillColor[i] = chart.fillColorFunc = getFillColor(chartObj, i, paneid, {
                        x1: 0,
                        x2: w,
                        y1: 0,
                        y2: h,
                        r: mathMin(w, h),
                        transform:
                            !layoutObj.isScalable &&
                            gradientOptions.type === 'radial' &&
                            gradientOptions.options.radial.gradientUnits !== 'objectBoundingBox'
                                ? `translate(${-transformation.translate[0]}, ${-transformation.translate[1]})`
                                : null
                    });
                    strokeColor[i] = getStrokeColor(chartObj, i);
                });

                container.attrs({
                    transform: `translate(${transformation.translate}) scale(${transformation.scale})`,
                    'text-anchor': MIDDLE_ALIGN
                });

                var boundBoxGroup = appendEle(container, 'g', [1], '.', 'class', 'boundBoxes'),
                    textGroup = appendEle(container, 'g', [1], '.', 'class', 'words');

                var boundBox = appendEle(boundBoxGroup, PATH_ELEMENT, layoutObj.hasBoundBox ? layoutObj.processedData : []);
                boundBox = boundBox.filter(filterAccessor);
                boundBox.each(function(d) {
                    if (wordCloudUtils.hasBoundBox(d)) {
                        _wordcloudBoundBoxSvg(d3_select(this), d, layoutObj, commonRendererProp);
                    }
                });

                var text = appendEle(textGroup, TEXT_ELEMENT, layoutObj.processedData);
                text = text.filter(filterAccessor);
                applyStaticStyle(container, commonPlotOptions);

                text.each(function(d) {
                    var ele = d3_select(this),
                        seriesIndex = _getSeriesIndex(d),
                        itemIndex = _getItemIndex(d),
                        layout = d.layout,
                        isMultiLine = wordCloudUtils.isMultiLine(layout),
                        plotOptions = data[seriesIndex].plotoptions,
                        plotLevelStrokeWidth = plotOptions && plotOptions.strokeColor,
                        _stroke = strokeColor[seriesIndex](
                            multiColoring ? itemIndex : seriesIndex,
                            d,
                            plotLevelStrokeWidth !== UNDEFINED ? plotLevelStrokeWidth : commonPlotOptions.strokeColor
                        ),
                        datalabelOptions = layoutObj.datalabelOptions[seriesIndex];

                    applyDatalabelFonts(ele, datalabelOptions, chartObj);
                    ele[layout.nodeList.contentType](layout.nodeList.innerSVG);

                    ele.styles({
                        'font-size': layout.size * fontScaleFactor + PX,
                        fill: fillColor[seriesIndex](multiColoring ? itemIndex : seriesIndex, d),
                        stroke: _stroke,
                        filter: NULL,
                        'stroke-linejoin': 'round'
                    }).attrs({
                        x: layout.x * fontScaleFactor,
                        y: layout.y * fontScaleFactor - (isMultiLine ? (layout.height * fontScaleFactor) / 2 : 0),
                        dy: `${wordCloudUtils.getDy(layout)}em`,
                        transform: `rotate(${radianToDegree(layout.rotation)} ${layout.x * fontScaleFactor} ${layout.y * fontScaleFactor})`,
                        id: `word_${seriesIndex}_${d.subSeriesIndex}_${itemIndex}`,
                        class: `zc_wordcloud series_${seriesIndex}`
                    });

                    ele.styles(
                        applyTextShadow4Firefox(
                            _stroke,
                            pick(plotOptions && plotOptions.strokeWidth, commonPlotOptions.strokeWidth),
                            datalabelOptions.textShadow
                        )
                    );

                    if (isMultiLine) {
                        translateTspan(ele, layout.x * fontScaleFactor);
                    }

                    if (plotOptions) {
                        applyStaticStyle(ele, plotOptions);
                    }
                });

                setPivotAttributes(text, chartObj);
                setOpacityForTickFilterElements(text, chartObj, 'wordcloud');

                //Animation
                if (!isredraw) {
                    text.style('opacity', initialOpacity).each(function(d) {
                        var layout = d.layout,
                            tX,
                            tY;

                        if (animationType === SIZING_ANIMATION) {
                            tX = dimension.width / 2;
                            tY = dimension.height / 2;
                        } else if (animationType === RANDOMOUT_ANIMATION) {
                            var usedWidth = w / scaleFactor,
                                usedHeight = h / scaleFactor;

                            tX =
                                mathRandom() * (usedWidth - layout.width * fontScaleFactor) -
                                dimension.x +
                                (layout.width * fontScaleFactor) / 2 +
                                (w - usedWidth) / 2;
                            tY =
                                mathRandom() * (usedHeight - layout.height * fontScaleFactor) -
                                dimension.y +
                                (layout.height * fontScaleFactor) / 2 +
                                (h - usedHeight) / 2;
                        } else {
                            tX = layout.x * fontScaleFactor;
                            tY = layout.y * fontScaleFactor;
                        }

                        d3_select(this).attrs({
                            x: tX,
                            y: tY - (wordCloudUtils.isMultiLine(layout) ? (layout.height * fontScaleFactor) / 2 : 0),
                            transform: `rotate(${radianToDegree(layout.rotation)} ${tX} ${tY})`
                        });
                    });
                }

                text.transition()
                    .duration(duration)
                    .delay(delay)
                    .ease(easingType)
                    .style(
                        'opacity',
                        isFacet(chartObj)
                            ? function() {
                                  return d3_select(this).attr('filtered-opacity') || 1;
                              }
                            : NULL
                    )
                    .attrs({
                        x: function(d) {
                            return d.layout.x * fontScaleFactor;
                        },
                        y: function(d) {
                            var layout = d.layout;
                            return (
                                layout.y * fontScaleFactor -
                                (wordCloudUtils.isMultiLine(layout) ? (layout.height * fontScaleFactor) / 2 : 0)
                            );
                        },
                        transform: function(d) {
                            var layout = d.layout;
                            return `rotate(${radianToDegree(layout.rotation)} ${layout.x * fontScaleFactor} ${layout.y * fontScaleFactor})`;
                        }
                    });

                chartObj.timerObj[chartObj.timerObj.length] = chart.renderedTimer = setTimeout(function() {
                    layoutObj.paneid = isFacet(chartObj) ? `${data[0].row}_${data[0].column}` : 0;

                    //Annotation preprocessing
                    if (!annotation._empty) {
                        data.forEach(function(series, i) {
                            notesArray = annotation.preprocessAnnotations(
                                data,
                                i,
                                function(dt, k, j) {
                                    var layout = dt.layout,
                                        x = transformation.translate[0] + (layout.x - layout.width / 2) * scaleFactor,
                                        y =
                                            transformation.translate[1] +
                                            (layout.y - layout.height * wordCloudUtils.getDy(layout)) * scaleFactor;

                                    return [x, y];
                                },
                                notesArray,
                                chartObj,
                                rp
                            );
                        });
                    }

                    chart.chartRendered = true;
                    isredraw = true;
                }, transitionDuration + 1000 / 60);
            }
        });
        return chart;
    }

    chart.commonRendererProp = function(_) {
        if (!arguments.length) {
            return commonRendererProp;
        }
        commonRendererProp = _;
        return chart;
    };

    chart.getHighlighter = function(chartObj, index) {
        return (chart.highlighter = new WordcloudHighlighter(chartObj, chart));
    };

    function applyStaticStyle(element, options) {
        element.styles({
            'paint-order': 'stroke',
            'fill-opacity': options.fillOpacity,
            'stroke-width': options.strokeWidth,
            'stroke-opacity': options.strokeOpacity
        });
    }

    return chart;
};

export var _wordcloudChart = svg_plot_renderer.wordcloud;
