// @zohocharts/zc/src/wordcloud/wordcloud-utils.js
//$Id$
//wordCloudUtils.js
import { validateNullVal, isObjectPropDefined, isUndefined } from '../utils/zcutils';
import { NULL } from '../core/zc-constants';
import { getMergedBoundBoxOptions, getBoundBoxStrokeWidth } from '../label/boundBoxUtils';

export var wordCloudUtils = {};

/**
 * @param {Array} d
 * @return {Number} seriesIndex
 */
wordCloudUtils.getSeriesIndex = function(d) {
    return d.seriesIndex;
};

/**
 * @param {Array} d
 * @return {Number} itemIndex
 */
wordCloudUtils.getItemIndex = function(d) {
    return d.itemIndex;
};

/**
 * @param {object} obj
 * @return {Number} dy
 */
wordCloudUtils.getDy = function(obj) {
    return wordCloudUtils.isMultiLine(obj) ? 0.9 : 0.35;
};

/**
 * @param {object} obj
 * @return {Boolean} isMultiLine
 */
wordCloudUtils.isMultiLine = function(obj) {
    return (obj.nodeList ? obj.nodeList.format : obj.format).multiLine;
};

/**
 * @param {object} layout
 * @param {object} datalabelOptions
 * @param {object} commonPlotOptions
 * @param {object} chartObj
 * @return {object} {x, y, width, heihgt, r}
 */
wordCloudUtils.getBoundBoxSize = function(layout, datalabelOptions, commonPlotOptions, chartObj, layoutObj) {
    var boundboxOptions = getMergedBoundBoxOptions(datalabelOptions, chartObj),
        padding = (validateNullVal(commonPlotOptions.padding, 2) + getBoundBoxStrokeWidth(datalabelOptions, chartObj)) / 2,
        scaleFactor = layoutObj.output.scaleFactor,
        fontScaleFactor = layoutObj.isScalable ? 1 : scaleFactor;

    return {
        x: layout.x * fontScaleFactor - (layout.width / 2) * fontScaleFactor + padding,
        y:
            layout.y * fontScaleFactor -
            (layout.height / 2) * fontScaleFactor +
            padding +
            (wordCloudUtils.isMultiLine(layout)
                ? layout.size * fontScaleFactor * wordCloudUtils.getDy(layout) - (layout.height / 2) * fontScaleFactor
                : 0),
        width: layout.width * fontScaleFactor - 2 * padding,
        height: layout.height * fontScaleFactor - 2 * padding,
        r: chartObj.dataObject.getTransRectValue('radius', NULL, NULL, NULL, NULL, boundboxOptions.radius)
    };
};

/**
 * @param {Object} layoutObj
 * @param {Object} dimension
 * @return {Object} {translate, scale}
 */
wordCloudUtils.getContainerTransformation = function(layoutObj, dimension) {
    var scaleFactor = layoutObj.output.scaleFactor,
        x = dimension.x - (dimension.width * (scaleFactor - 1)) / 2,
        y = dimension.y - (dimension.height * (scaleFactor - 1)) / 2;

    return {
        translate: [x, y],
        scale: layoutObj.isScalable ? scaleFactor : 1
    };
};

/**
 * @param  {Object} d
 * @return {Boolean}
 */
wordCloudUtils.hasBoundBox = function(d) {
    var innerSVG = isObjectPropDefined(d, 'layout.nodeList.innerSVG', true);
    return d.layout.size && !(isUndefined(innerSVG) || innerSVG === '');
};
