// @zohocharts/zc/src/wordcloud/wordcloud-layout.js
//$Id$
//wordcloudLayout.js
import { DataProcessor } from '../core/data-processor';
import { Scale } from '../scale/scale-component';
import { fontScaleParser } from './font-scale-parser';
import { wordCloudUtils } from './wordcloud-utils';
import { NULL } from '../core/zc-constants';
import { PERCENT_SHOWAS } from '../label/showas-constants';
import { $Browser } from '../core/browser-detect';
import { getSVGsymbol } from '../utils/svg-symbols';
import { DATALABELS_LABELTYPE } from '../label/labeltype-constants';
import { rendererBoundUtils } from '../plot/renderer-bound-utils';
import { validateNullVal, defined, isUndefined, isArray, mergeJSON, arrayIncludes, isObjectPropDefined } from '../utils/zcutils';
import { WordCloud } from './wordcloud';
import { applyDatalabelFonts, getMergedDatalabelOptions } from '../label/datalabelsUtils';
import { dataLabelTextContent } from '../label/datalabel-content';
import { isBoundBoxEnabled, getBoundBoxPadding, getBoundBoxStrokeWidth } from '../label/boundBoxUtils';
import { geometryUtils } from '../utils/geometry-utils';
import { createElement } from '../utils/zcsvgutils';
import { btoa } from '../core/zcglobal';
import { convertNull } from '../core/data-processor/size/convert-null';

export var wordcloudLayout = function(commonRendererProp, data) {
    var rp,
        chartObj,
        commonPlotOptions,
        processedData,
        dimension,
        sum,
        fontScale,
        datalabelOptions,
        GLT,
        fontFamily,
        fontStyle,
        fontWeight,
        yValues,
        hasBoundBox,
        globalRotation,
        globalRotationCount,
        seriesSpecificRotation,
        seriesSpecificRotationCount,
        _getSeriesIndex = wordCloudUtils.getSeriesIndex,
        getValue = function(d) {
            var yaxiscolumnorder = DataProcessor.helpers.getParsedYAxisOrder(data[_getSeriesIndex(d)].yaxiscolumnorder);
            return chartObj.dataset.getY(d, yaxiscolumnorder);
        },
        paneid = `${data[0].row}_${data[0].column}`,
        isScalable = true,
        fontData = [],
        processData = function() {
            data.forEach(function(series, seriesIndex) {
                let plotOptions = series.plotoptions;
                let gradientOptions =
                        plotOptions && plotOptions.gradients
                            ? mergeJSON(commonPlotOptions.gradients, plotOptions.gradients, true)
                            : commonPlotOptions.gradients,
                    activeData = [];
                isScalable = isScalable && !arrayIncludes(['linear', 'radial'], gradientOptions.type);

                if (!series.disabled) {
                    series.data.forEach(function(subseries, subSeriesIndex) {
                        subseries.forEach(function(d, itemIndex) {
                            d.seriesIndex = seriesIndex;
                            d.subSeriesIndex = subSeriesIndex;
                            d.itemIndex = itemIndex;
                            d.layout = NULL;
                            d.paneid = paneid;
                            d.level = `${seriesIndex}_${subSeriesIndex}_${itemIndex}`; //Hack because of double click event return wrong seriesIndex, itemIndex

                            var yVal = getValue(d);

                            if (defined(commonPlotOptions.nullAs) || defined(yVal)) {
                                //Omit null|undefined values
                                sum += yVal;
                                yValues.push(yVal);
                                processedData.push(d);
                                activeData.push(d);
                            }
                        });
                    });
                }
                fontData.push({ data: [activeData] });
                // chartObj.dataObject.saveEncode('y', 0, seriesIndex, { data: [activeData] }); // For font-scale
            });
        },
        setTextProperties = function() {
            var percentFunc = function(d, showAs) {
                return showAs === PERCENT_SHOWAS ? sum : fontScale.domain()[1];
            };

            data.forEach(function(series, seriesIndex) {
                var options = getMergedDatalabelOptions(commonPlotOptions, series.plotoptions, chartObj, seriesIndex, paneid);
                var font = applyDatalabelFonts(NULL, options, chartObj);

                fontFamily.push(font['font-family']);
                fontStyle.push(font['font-style']);
                fontWeight.push(font['font-weight']);

                datalabelOptions.push(options);
                GLT.push(dataLabelTextContent(chartObj, rp, options, DATALABELS_LABELTYPE, percentFunc, seriesIndex, undefined, paneid));
                hasBoundBox = hasBoundBox || isBoundBoxEnabled(options, chartObj);

                var rotation = series.plotoptions && series.plotoptions.rotation;
                if (defined(rotation)) {
                    seriesSpecificRotation[seriesIndex] = !isArray(rotation) ? [rotation] : rotation;
                    seriesSpecificRotationCount[seriesIndex] = 0;
                }
            });
        },
        getMask = function() {
            var maskOptions = commonPlotOptions.mask,
                htmlEl = !$Browser.IE && !$Browser.PHANTOMJS && maskOptions.htmlEl,
                imageUrl = maskOptions.imageUrl,
                symbol = maskOptions.symbol,
                mask;

            if (htmlEl) {
                mask = `data:image/svg+xml;${$Browser.IE ? 'base64' : 'utf8'},${
                    $Browser.IE ? btoa(unescape(encodeURIComponent(htmlEl))) : htmlEl
                }`;
            } else if (imageUrl) {
                mask = imageUrl; // "https://upload.wikimedia.org/wikipedia/commons/thumb/d/d0/Heart_font_awesome.svg/1024px-Heart_font_awesome.svg.png", //check with http, relative image
            } else if (symbol) {
                mask = getSVGsymbol(symbol);
                mask._name = symbol;
            }
            return mask;
        },
        drawLayout = function(render) {
            var self = this,
                maskOptions = commonPlotOptions.mask,
                dataLabels = commonPlotOptions.datalabels;

            var options = {
                width: dimension.width,
                height: dimension.height,
                rotation: function(d) {
                    var seriesIndex = _getSeriesIndex(d),
                        degree;

                    if (seriesSpecificRotation[seriesIndex]) {
                        degree =
                            seriesSpecificRotation[seriesIndex][
                                seriesSpecificRotationCount[seriesIndex]++ % seriesSpecificRotation[seriesIndex].length
                            ];
                    } else {
                        degree = globalRotation[globalRotationCount++ % globalRotation.length];
                    }

                    return geometryUtils.degreeToRadian(degree);
                },
                fontSize: function(d) {
                    var value = getValue(d);
                    if (defined(commonPlotOptions.nullAs) && isUndefined(value)) {
                        let nullAsOptions = {
                            config: commonPlotOptions,
                            scale: fontScale,
                            columnKey: chartObj.dataset.getColumnIdx('y', 0)
                        };
                        value = convertNull(value, nullAsOptions);
                    }

                    return defined(value) ? fontScale(value) : 0;
                },
                fontFamily: function(d) {
                    return fontFamily[_getSeriesIndex(d)];
                },
                fontStyle: function(d) {
                    return fontStyle[_getSeriesIndex(d)] || 'normal';
                },
                fontWeight: function(d) {
                    return fontWeight[_getSeriesIndex(d)] || 'normal';
                },
                content: function(d) {
                    var yVal = getValue(d),
                        format = GLT[_getSeriesIndex(d)];

                    return {
                        type: format.contentType,
                        value: format(d),
                        format: format
                    };
                },
                padding: function(d) {
                    var currentDatalabelOptions = datalabelOptions[_getSeriesIndex(d)],
                        padding = validateNullVal(commonPlotOptions.padding, 2),
                        boundBoxPadding = 0;

                    if (isBoundBoxEnabled(currentDatalabelOptions, chartObj)) {
                        boundBoxPadding =
                            getBoundBoxPadding(currentDatalabelOptions, chartObj) * 2 +
                            getBoundBoxStrokeWidth(currentDatalabelOptions, chartObj);
                    }

                    return padding + boundBoxPadding;
                },
                strokeWidth: function(d) {
                    var plotOptions = data[_getSeriesIndex(d)].plotoptions,
                        defaultStrokeWidth =
                            plotOptions && defined(plotOptions.strokeWidth) ? plotOptions.strokeWidth : commonPlotOptions.strokeWidth;
                    return validateNullVal(defaultStrokeWidth, 1);
                },
                onload: function(args) {
                    chartObj.hideLoading(chartObj.container);
                    chartObj.svg.selectAll('g.zc-wordcloud-dummy').remove();
                    self.output = args;
                    render();
                },
                origin: NULL, //have to work
                mask: getMask(),
                maskBackground: maskOptions.background.color,
                invert: maskOptions.invert,
                textOverflow: dataLabels.textOverflow || 'ellipsis',
                textOverlap: dataLabels.textOverlap || 'adjust', //hidden | adjust
                maxWidth: dataLabels.maxWidth,
                fontScale: fontScale,
                container:
                    isObjectPropDefined(dataLabels, 'preCalculation.singleNode', true) === false
                        ? createElement('g', '.zc-wordcloud-dummy', [1], chartObj.svg).style('visibility', 'hidden')
                        : chartObj.getPatternSvg(),
                size:
                    commonPlotOptions.size === 'auto'
                        ? {}
                        : {
                              min: commonPlotOptions.minSize,
                              max: commonPlotOptions.maxSize
                          },
                boundBox: hasBoundBox,
                maxSizeFactor: commonPlotOptions.maxSizeFactor,
                preCalculation: dataLabels.preCalculation
            };

            //Todo: Safari breathing time
            // setTimeout(function() {
            WordCloud(processedData, options);
            // }, 0); //Safari: for breathing time. without this timeout loader will not be shown. (e.g) wordcloud-10.json
        },
        initVariables = function() {
            rp = commonRendererProp.params;
            chartObj = rp.chartObj;
            commonPlotOptions = rp.commonPlotOptions;
            dimension = rendererBoundUtils.getAvailableBound(rp.bound, commonPlotOptions);

            sum = 0;
            processedData = [];

            datalabelOptions = [];
            GLT = [];
            fontFamily = [];
            fontStyle = [];
            fontWeight = [];
            yValues = [];

            seriesSpecificRotation = [];
            seriesSpecificRotationCount = [];

            globalRotation = commonPlotOptions.rotation || [0];
            globalRotation = !isArray(globalRotation) ? [globalRotation] : globalRotation;
            globalRotationCount = 0;
        },
        createFontScale = () => {
            let config = { ...commonPlotOptions, _data: fontData },
                { ranges, stops, minRange, maxRange, base, nullAs } = config,
                unknown = 0,
                parser = fontScaleParser('y', 0, chartObj, config),
                scaleComponent = new Scale('y', 0, { ranges, stops, minRange, maxRange, base, nullAs, unknown, ...parser }, chartObj); // TODO: check with config
            return scaleComponent.getScale();
        },
        init = function() {
            initVariables();
            chartObj.showLoading(chartObj.container);
            processData();
            fontScale = createFontScale();
            setTextProperties();
        };

    init();

    return {
        rp,
        processedData,
        dimension,
        sum,
        fontScale,
        datalabelOptions,
        fontFamily,
        fontStyle,
        fontWeight,
        drawLayout,
        hasBoundBox,
        data,
        isScalable: $Browser.PHANTOMJS && chartObj.plot.renderer.mode !== 'canvas' ? isScalable : true
    };
};
