// @zohocharts/zc/src/wordcloud/wordcloud-highlight.js
import { BarHighlighter } from '../bar/bar-highlight';
import { TEXT_ELEMENT } from '../core/element-constants';
import { MIDDLE_ALIGN } from '../core/align-constants';
import { HIGHLIGHT_CONTEXT } from '../events/visualEffects/highlight/highlight-context';
import { NULL, UNDEFINED, PX, mathMin } from '../core/zc-constants';
import { createElement } from '../utils/zcsvgutils';
import { inherit } from '../utils/zcutils';
import { euclideanDistance } from '../utils/geometry/_point';
import { isMultiColoring, getRawColor, getFillColor, getStrokeColor } from '../utils/colorutils';
import { isPointVisible } from '../events/events-utils';
import { radianToDegree } from '../polaraxis/polarutils';
import { applyDatalabelFonts } from '../label/datalabelsUtils';
import { isFacet } from '../axis/hierarchy-axis-utils';
import { wordCloudUtils } from './wordcloud-utils';
import { patternFillEffect } from '../events/visualEffects/highlight/pattern-fill';
import { getRendererIndex } from '../core/data-parser';
import { translateTspan } from '../label/tspan-utils';

/*
 Doubts:
 1) Whether sharedseries will be applicable?
 Pending:
 */
export function WordcloudHighlighter(chart, renderer) {
    this.chart = chart;
    this.renderer = renderer;
}

inherit(WordcloudHighlighter, BarHighlighter);
let proto = WordcloudHighlighter.prototype;

proto._defaultH = function() {
    return [
        {
            shape: TEXT_ELEMENT,
            path: this.path,
            data: this.data,
            context: this.renderingContext,
            translate: this.translateContext,
            styles: patternFillEffect
        }
    ];
};

proto.renderingContext = function() {
    return HIGHLIGHT_CONTEXT.get('nonaxis').call(this, this.chart);
};

proto.translateContext = function(point) {
    let chart = this.chart,
        renderer = getRenderer(point, chart, this.renderer),
        bound = renderer.commonRendererProp().bound(),
        layout = renderer.layout,
        dimension = layout.dimension,
        transformation = wordCloudUtils.getContainerTransformation(layout, dimension),
        translate = transformation.translate,
        plotArea = chart.plotarea;

    return [plotArea.left + translate[0] + bound.x, plotArea.top + translate[1] + bound.y, transformation.scale];
};

proto.data = function(point) {
    let chart = this.chart,
        renderer = getRenderer(point, chart, this.renderer),
        layout = renderer.layout,
        dimension = layout.dimension,
        bound = renderer.commonRendererProp().bound();

    return function() {
        if (!chart.cache.data[point._key]) {
            chart.cache.data[point._key] = {};
        }

        let cache = chart.cache.data[point._key];
        cache._point = point;
        let tileBound = point.layout;

        if (!tileBound) {
            cache._visible = point._visible = false;
            return point;
        }

        let transformation = wordCloudUtils.getContainerTransformation(layout, dimension).translate;
        let scaleFactor = layout.output.scaleFactor;
        let { x, y, width, height, rotation } = point.layout;
        x *= scaleFactor;
        y *= scaleFactor;
        width *= scaleFactor;
        height *= scaleFactor;

        cache._x = point._x = cache._x || x + transformation[0] + bound.x;
        cache._y = point._y = cache._y || y + transformation[1] + bound.y;
        cache._width = point._width = cache._width || width;
        cache._height = point._height = cache._height || height;
        cache._centric = point._centric = cache._centric || true;
        cache._rotation = point._rotation = cache._rotation || rotation;
        cache._size = point._size =
            2 * euclideanDistance({ x: point._x, y: point._y }, { x: point._x - width / 2, y: point._y - height / 2 });
        cache._uid = point._uid = chart.dataset.getX(point);
        cache._visible = point._visible = isPointVisible(chart, point);

        return point;
    };
};

proto.options = function(point) {
    let { seriesIndex, itemIndex, paneid } = point,
        chart = this.chart,
        renderer = getRenderer(point, chart, this.renderer),
        rp = renderer.commonRendererProp().params,
        multiColoring = isMultiColoring(rp.commonPlotOptions, chart),
        layout = renderer.layout,
        dimension = layout.dimension,
        //
        scaleFactor = layout.output.scaleFactor,
        fontScaleFactor = layout.isScalable ? 1 : scaleFactor,
        rawColor = getRawColor(chart, multiColoring ? itemIndex : seriesIndex, point, seriesIndex, paneid);

    let textInfo = point.layout,
        isMultiLine = wordCloudUtils.isMultiLine(textInfo),
        highlightX = textInfo.x * fontScaleFactor,
        highlightY = textInfo.y * fontScaleFactor,
        plotLevelStrokeWidth,
        w = dimension.width,
        h = dimension.height,
        _fillColor,
        _strokeColor;

    if (isFacet(chart)) {
        _fillColor = getFillColor(chart, seriesIndex, paneid, { x1: 0, x2: w, y1: 0, y2: h, r: mathMin(w, h) }, NULL, 'svg');
        _strokeColor = getStrokeColor(chart, seriesIndex, paneid);
    } else {
        _fillColor = getFillColor(chart, 0, null, { x1: 0, x2: w, y1: 0, y2: h, r: mathMin(w, h) }, NULL, 'svg');
        _strokeColor = getStrokeColor(chart, 0);
    }

    return {
        x: highlightX,
        y: highlightY - (isMultiLine ? (textInfo.height * fontScaleFactor) / 2 : 0),
        dy: wordCloudUtils.getDy(textInfo),
        rotation: [radianToDegree(textInfo.rotation), highlightX, highlightY],
        fontSize: textInfo.size * fontScaleFactor,
        font: layout.datalabelOptions[seriesIndex],
        contentType: textInfo.nodeList.contentType,
        innerSVG: textInfo.nodeList.innerSVG,
        color: rawColor,
        fillColor: _fillColor(multiColoring ? itemIndex : seriesIndex, point),
        strokeColor: _strokeColor(
            multiColoring ? itemIndex : seriesIndex,
            point,
            plotLevelStrokeWidth !== UNDEFINED ? plotLevelStrokeWidth : rp.commonPlotOptions.strokeColor
        )
    };
};

proto.path = function(point, model) {
    let self = this,
        context = model.context.call(self),
        translate = model.translate.call(self, point),
        options = this.options(point),
        { x, y, dy, rotation, fontSize, font, contentType, innerSVG } = options,
        textInfo = point.layout,
        isMultiLine = wordCloudUtils.isMultiLine(textInfo),
        chart = this.chart;

    context.style('display', '').style('opacity', null);

    let groupContext = createElement(
        'g',
        `.wordcloud-highlight${point.paneid}${point.seriesIndex}${point.subSeriesIndex}`,
        [point],
        context
    );
    groupContext.attr('transform', `translate(${translate[0]},${translate[1]}) scale(${translate[2]})`);
    let textContext = createElement(TEXT_ELEMENT, `.wordcloud-highlight`, [point], groupContext);
    applyDatalabelFonts(textContext, font, chart);
    textContext[contentType](innerSVG);
    textContext
        .attrs({
            x,
            y,
            dy: `${dy}em`,
            transform: `rotate(${rotation[0]} ${rotation[1]} ${rotation[2]})`,
            'text-anchor': MIDDLE_ALIGN
        })
        .styles({
            'font-size': fontSize + PX,
            'paint-order': 'stroke',
            'stroke-linejoin': 'round'
        });

    if (isMultiLine) {
        translateTspan(textContext, textInfo.x);
    }

    return textContext;
};

function getRenderer(point, chart, renderer) {
    return point ? chart.renderer[getRendererIndex(point, chart)] : renderer;
}
