// @zohocharts/zc/src/wordcloud/wordcloud.js
//$Id$
//wordcloud.js
import { Registry } from '../core/zc-registry';
import { $Browser } from '../core/browser-detect';
import {
    mathMin,
    mathMax,
    mathFloor,
    mathCos,
    mathSin,
    mathRound,
    mathCeil,
    mathSQRT,
    mathPI,
    mathAbs,
    mathPow
} from '../core/zc-constants';
import { defined, bbox, perToPx, pick, objectValues } from '../utils/zcutils';
import { applyFont } from '../utils/zcsvgutils';
import { TEXT_ELEMENT, TSPAN_ELEMENT } from '../core/element-constants';
import { wordCloudUtils } from './wordcloud-utils';
import { CENTER_ALIGN, MIDDLE_ALIGN } from '../core/align-constants';
import { geometryUtils } from '../utils/geometry-utils';
import { naiveEllipsis, appendEle } from '../utils/zcsvgutils';
import { getSVGsymbol } from '../utils/svg-symbols';
import { radianToDegree } from '../polaraxis/polarutils';
import { translateTspan, getTspanBound } from '../label/tspan-utils';
import { getDefaultBoundingClientRect, getTspanBoundingClientRect } from '../axis/textBound';
import { NULL, PX } from '../core/zc-constants';
import { getUserFontValue } from '../label/font-helpers';
import { canvasUtils } from '../utils/canvasutils';
import { d3_rgb, d3_scaleLinear, d3_scalePoint } from '../core/d3-constants';

// https://github.com/timdream/wordcloud2.js/blob/gh-pages/API.md
var TRUE = true,
    FALSE = false,
    maxCanvasSize = $Browser.CHROME || $Browser.FIREFOX ? 32000 : $Browser.IE && $Browser.isTouchDevice ? 4000 : 8000,
    createCanvas = function() {
        return document.createElement('canvas');
    },
    createContext = function(canvas) {
        return canvas.getContext('2d');
    },
    isTspan = function(node) {
        return node.nodeName === TSPAN_ELEMENT;
    };

export var WordCloud = function(data, config) {
    var plotWidth = config.width,
        plotHeight = config.height,
        textOverflow = config.textOverflow,
        textOverlap = config.textOverlap,
        mask = config.mask,
        invert = config.invert || FALSE,
        fontScale = config.fontScale,
        maxWidth = defined(config.maxWidth) ? perToPx(config.maxWidth, plotWidth) : NULL,
        _fontFamily = config.fontFamily,
        _fontStyle = config.fontStyle,
        _fontWeight = config.fontWeight,
        _fontSize = config.fontSize,
        _content = config.content,
        _rotation = config.rotation,
        patternSvg = config.container,
        { maxSizeFactor } = config,
        pointsAtRadius = [],
        g = mathMin(8, mathMax(4, mathRound((16 * mathMin(plotWidth, plotHeight)) / 1024))), // Make sure gridSize is a whole number and is not smaller than 4px
        grid = [], // 2d array containing filling information (current canvas state)
        rotationCollection = [],
        outside = [0, 0],
        meta,
        maskImageData,
        areaOfMask,
        imageDataX,
        imageDataY,
        imageTotalWidth,
        imageTotalHeight,
        currentRowHeight,
        canvasCount,
        textImageData,
        drawnData,
        calculatedPosition,
        canvasMeta,
        drawnBound,
        { singleNode = true, timeOut = 0, adjust } = config.preCalculation || {},
        breathNInvoke = !singleNode ? func => setTimeout(func, timeOut) : func => func(),
        //
        dummyText = (function() {
            if (singleNode) {
                return appendEle(patternSvg, TEXT_ELEMENT, [1], '.', 'class', 'dummyText');
            } else {
                return appendEle(patternSvg, TEXT_ELEMENT, data, '.', 'class', 'dummyWord').attr('id', (d, i) => 'dummyWord_' + i);
            }
        })(),
        getDummyText = i => (singleNode ? dummyText : patternSvg.selectAll('text#dummyWord_' + i)),
        // width and height of the grid
        ngx = mathCeil(plotWidth / g),
        ngy = mathCeil(plotHeight / g),
        fx = mathMin(1, ngx / ngy),
        fy = mathMin(1, ngy / ngx),
        center = (center = config.origin) ? [center[0] / g, center[1] / g] : [ngx / 2, ngy / 2], // position of the center of the cloud
        maxRadius = mathFloor(mathSQRT(ngx * ngx + ngy * ngy)), // Maxium radius to look for space
        noOfPointAtRadius = data.length < 100 ? 8 : 1,
        /* Get points on the grid for a given radius away from the center */
        getPointsAtRadius = function(radius) {
            if (pointsAtRadius[radius]) {
                return pointsAtRadius[radius];
            }

            var T = radius * noOfPointAtRadius, // Look for these number of points on each radius
                t = T,
                points = []; // Getting all the points at this radius

            if (radius === 0) {
                points.push([center[0], center[1] + outside[1], 0]);
            }

            while (t--) {
                // distort the radius to put the cloud in shape
                var theta = (t / T) * 2 * mathPI;

                // Push [x, y, t]; t is used solely for getTextColor()
                points.push([center[0] + radius * mathCos(-theta) * fx, center[1] + outside[1] + radius * mathSin(-theta) * fy, theta]);
            }

            return (pointsAtRadius[radius] = points);
        },
        getCanvas = function(d) {
            var canvas = createCanvas();
            canvas.setAttribute('width', d.width);
            canvas.setAttribute('height', d.height);
            d.context = createContext(canvas);
        },
        getContext = function(arg) {
            var index = arg.constructor === Object ? arg.canvasIndex : arg;
            return canvasMeta[index].context;
        },
        readImageData = function(d, i) {
            textImageData[i] = getContext(i).getImageData(0, 0, d.width, d.height).data; // Get the pixels of all the words
        },
        getWordPadding = function(d) {
            return config.strokeWidth(d) + config.padding(d);
        },
        drawText = function(d, info) {
            // XXX: We cannot because textBaseline = 'top' here because
            // Firefox and Chrome uses different default line-height for canvas.
            // Please read https://bugzil.la/737852#c6.
            // Here, we use textBaseline = 'middle' and draw the text at exactly
            // 0.5 * fontSize lower.
            var fontSize = info.size,
                globalFont = getFont(d),
                bb = info.bb,
                nodeList = info.nodeList,
                dy = wordCloudUtils.getDy(nodeList),
                x = 0,
                y = 0,
                context = getContext(info);

            context.strokeStyle = '#000';
            context.lineWidth = getWordPadding(d);
            context.lineJoin = 'bevel'; // ZC1459 // add this too context.miterLimit=2;
            context.fillStyle = '#000';

            if (nodeList.contentType === 'html') {
                context.textAlign = CENTER_ALIGN;
                nodeList.forEach(function(a) {
                    var userStyle = a.style,
                        font,
                        fSize;

                    if (userStyle.length) {
                        font = getUserFontValue(applyFont(NULL, [getFontStyles(d)]), userStyle);
                        fSize = parseFloat(userStyle['font-size']);
                    }

                    x = a.dx || 0;
                    y = (a.dy || 0) * fontSize + fontSize * dy - (wordCloudUtils.isMultiLine(nodeList) ? bb.height / 2 : 0);

                    context.font = font || globalFont;
                    context.strokeText(a.text, x, y);
                    context.fillText(a.text, x, y);
                });
            } else {
                y = fontSize * dy;
                context.textAlign = CENTER_ALIGN;
                context.font = globalFont;
                context.strokeText(nodeList.innerSVG, x, y);
                context.fillText(nodeList.innerSVG, x, y);
            }
        },
        getFontStyles = function(d) {
            return {
                fontStyle: _fontStyle(d),
                fontWeight: _fontWeight(d),
                fontSize: _fontSize(d),
                fontFamily: _fontFamily(d)
            };
        },
        getFont = function(d) {
            return `${_fontStyle(d)} ` + `normal` + ` ${_fontWeight(d)} ${_fontSize(d).toString(10)}${PX} ${_fontFamily(d)}`;
        },
        sortByFontSizeAccesor = function(a, b) {
            return _fontSize(b) - _fontSize(a);
        },
        sortByContentAccesor = function(a, b) {
            return _content(b).value.length - _content(a).value.length;
        },
        updateBound = function(bound, start, end) {
            bound[0] = mathMin(bound[0], start[0]);
            bound[1] = mathMin(bound[1], start[1]);
            bound[2] = mathMax(bound[2], end[0]);
            bound[3] = mathMax(bound[3], end[1]);
        },
        getOccupiedArrayNBound = function(info) {
            // Read the pixels and save the information to the occupied array
            var cgw = info.gw,
                cgh = info.gh,
                cx = info.canvasBound[0],
                cy = info.canvasBound[1],
                canvasIndex = info.canvasIndex,
                width = canvasMeta[canvasIndex].width,
                imageData = textImageData[canvasIndex],
                occupied = [],
                bound = [cgw / 2, cgh / 2, cgw / 2, cgh / 2],
                gx = cgw,
                gy,
                x,
                y;

            while (gx--) {
                gy = cgh;
                while (gy--) {
                    y = g;
                    singleGridLoop: {
                        while (y--) {
                            x = g;
                            while (x--) {
                                if (imageData[((gy * g + y + cy) * width + (gx * g + x + cx)) * 4 + 3]) {
                                    occupied.push([gx, gy]);
                                    updateBound(bound, [gx, gy], [gx, gy]);
                                    break singleGridLoop;
                                }
                            }
                        }
                    }
                }
            }

            return {
                occupied: occupied,
                bound: bound
            };
        },
        wordOverflow = function(ele, d, deg, customWidth) {
            var textNode = ele.node(),
                nodes = textNode.childElementCount
                    ? ele
                          .selectAll('*')
                          .filter(function() {
                              return isTspan(this) && !this.childElementCount;
                          })
                          .nodes()
                    : [textNode],
                degree_360 = radianToDegree(deg) % 360,
                degree_90 = degree_360 > 90 ? 180 - degree_360 : degree_360;

            ele.attr('dy', '1em').attr('transform', `rotate(${degree_90} 0 0)`);

            var overflowPattern = textOverflow === 'ellipsis' ? '..' : '',
                overflowPatternLength = overflowPattern.length,
                textNode_bb = getDefaultBoundingClientRect(textNode),
                padding = config.padding(d),
                w = (defined(customWidth) ? customWidth : plotWidth) - padding - (adjust ? g : 0), // ZC1459: consider padding
                h = plotHeight - padding,
                available = {
                    left: textNode_bb.left,
                    top: textNode_bb.top,
                    right: textNode_bb.left + w,
                    bottom: textNode_bb.top + h,
                    width: w,
                    height: h
                },
                getTextBound_wordcloud = function(node) {
                    var clientRect = isTspan(node) ? getTspanBoundingClientRect : getDefaultBoundingClientRect,
                        bb = clientRect(node, 1);

                    if (isTspan(node)) {
                        availableWidth = available.width - mathMax(0, bb.left - available.left) - mathMax(0, available.right - bb.right);
                        availableHeight = available.height - mathMax(0, bb.top - available.top) - mathMax(0, available.bottom - bb.bottom);
                    }

                    return {
                        width: bb.width,
                        height: bb.height
                    };
                },
                overflowLength,
                croppedTextLength,
                availableWidth,
                availableHeight,
                textBound;

            nodes.forEach(function(node) {
                availableWidth = w;
                availableHeight = h;
                textBound = getTextBound_wordcloud(node);

                if (textBound.width - availableWidth > 1) {
                    // Test_941.json, Test_942.json
                    overflowLength = (textBound.width - availableWidth) / textBound.width;
                    croppedTextLength = mathCeil((1 - overflowLength) * node.textContent.length);
                    node.textContent = node.textContent.substring(0, croppedTextLength - overflowPatternLength) + overflowPattern;
                    textBound = getTextBound_wordcloud(node);
                }

                if (textBound.height - availableHeight > 1) {
                    overflowLength = (textBound.height - availableHeight) / textBound.height;
                    croppedTextLength = mathCeil((1 - overflowLength) * node.textContent.length);
                    node.textContent = node.textContent.substring(0, croppedTextLength - overflowPatternLength) + overflowPattern;
                    textBound = getTextBound_wordcloud(node);
                }

                if (textBound.width - availableWidth > 1 || textBound.height - availableHeight > 1) {
                    naiveEllipsis(node, node, node.textContent, textBound.width, textBound.height, textBound, overflowPattern);
                }

                if (node.textContent === overflowPattern) {
                    node.textContent = '';
                }
            });
        },
        drawWordsInSvg = function(d, i, customFontSize) {
            var content = _content(d),
                innerSVG = content.value,
                contentType = content.type,
                rotateDeg = rotationCollection[i],
                textEle = getDummyText(i);

            textEle[contentType](innerSVG);
            applyFont(textEle, [getFontStyles(d)]);
            if (customFontSize) {
                textEle.style('font-size', customFontSize + PX);
            }
            if (wordCloudUtils.isMultiLine(content)) {
                translateTspan(textEle, 0);
            }

            if (!customFontSize && textOverflow !== 'none') {
                if (maxWidth) {
                    wordOverflow(textEle, d, 0, maxWidth); //Have to enhance
                }
                wordOverflow(textEle, d, rotateDeg);
            }

            textEle
                .attr('text-anchor', MIDDLE_ALIGN)
                .attr('dy', `${wordCloudUtils.getDy(content)}em`)
                .attr('transform', `rotate(${radianToDegree(rotateDeg)} 0 0)`);
        },
        getTextInfo = function(d, i) {
            var fontSize = _fontSize(d),
                content = _content(d),
                padding = getWordPadding(d),
                contentType = content.type,
                rotateDeg = rotationCollection[i],
                nodeList = [];

            var textEle = getDummyText(i),
                bb = bbox(textEle.node()),
                boxWidth = bb.width + padding,
                boxHeight = bb.height + padding,
                // Calculate the actual dimension of the canvas, considering the rotation.
                cgh = mathCeil((boxWidth * mathAbs(mathSin(rotateDeg)) + boxHeight * mathAbs(mathCos(rotateDeg))) / g),
                cgw = mathCeil((boxWidth * mathAbs(mathCos(rotateDeg)) + boxHeight * mathAbs(mathSin(rotateDeg))) / g),
                width = cgw * g,
                height = cgh * g;

            nodeList.innerSVG = textEle[contentType]().trim(); // Why trim? If text has trailing spaces, wordcloud will overlap
            nodeList.contentType = contentType;
            nodeList.format = content.format;
            if (contentType === 'html') {
                var tspanInfo = getTspanBound(textEle, fontSize);
                Array.prototype.push.apply(nodeList, tspanInfo);
            }

            // Return information needed to create the text on the real canvas
            var info = {
                gw: cgw,
                gh: cgh,
                width: boxWidth,
                height: boxHeight,
                size: fontSize,
                rotation: rotateDeg,
                nodeList: nodeList,
                bb: bb // insteadof bb.height, can we use boxheight?????????
            };

            if (imageDataX + width > maxCanvasSize) {
                imageDataX = 0;
                imageDataY = imageDataY + currentRowHeight;
                currentRowHeight = 0;
            }

            if (imageDataY + height > maxCanvasSize) {
                canvasMeta[++canvasCount] = { width: 1, height: 1 };
                imageDataX = imageDataY = currentRowHeight = 0;
            }

            info.canvasBound = [imageDataX, imageDataY, width, height];
            info.canvasIndex = canvasCount;

            imageDataX += width;
            currentRowHeight = mathMax(currentRowHeight, height);
            canvasMeta[canvasCount].width = mathMax(canvasMeta[canvasCount].width, imageDataX);
            canvasMeta[canvasCount].height = imageDataY + currentRowHeight;

            return info;
        },
        drawWordsInCanvas = function(d, i) {
            var info = meta[i];
            if (!info) {
                return;
            }

            var context = getContext(info),
                canvasBound = info.canvasBound;
            context.save();
            context.translate(canvasBound[0] + canvasBound[2] / 2, canvasBound[1] + canvasBound[3] / 2);
            context.rotate(info.rotation);

            if (config.boundBox) {
                context.fillRect(-info.width / 2, -info.height / 2, info.width, info.height);
            } else {
                drawText(d, info);
            }

            context.restore();
        },
        isWithin = function(x, y) {
            return x >= ngx || y >= ngy || x < 0 || y < 0;
        },
        /* Determine if there is room available in the given dimension */
        canFitText = function(gx, gy, occupied) {
            // Go through the occupied points,
            // return false if the space is not available.

            return !occupied.some(function(d) {
                var px = gx + d[0],
                    py = gy + d[1];

                if (isWithin(px, py) || !grid[px][py]) {
                    return TRUE;
                }
            });
        },
        /* Fill the grid with empty state */
        createGrid = function() {
            var gx = ngx,
                gy,
                x,
                y;

            while (gx--) {
                grid[gx] = [];
                gy = ngy;
                while (gy--) {
                    if (!mask) {
                        grid[gx][gy] = TRUE;
                    } else {
                        grid[gx][gy] = !invert;
                        y = g;
                        singleGridLoop: while (y--) {
                            x = g;
                            while (x--) {
                                if (!maskImageData[((gy * g + y) * ngx * g + (gx * g + x)) * 4 + 3]) {
                                    grid[gx][gy] = invert;
                                    break singleGridLoop;
                                }
                            }
                        }
                    }
                }
            }
        },
        /* Update the filling information of the given space with occupied points.
               Draw the mask on the canvas if necessary. */
        updateGrid = function(position, info) {
            info.occupied.forEach(function(d) {
                var px = position[0] + d[0],
                    py = position[1] + d[1];

                if (!isWithin(px, py)) {
                    grid[px][py] = FALSE;
                }
            });
        },
        tryToPutWordAtPoint = function(gxy) {
            var info = this,
                gx = mathFloor(gxy[0] - info.gw / 2),
                gy = mathFloor(gxy[1] - info.gh / 2 + outside[1]);

            // If we cannot fit the text at this position, return false
            // and go to the next position.
            if (!canFitText(gx, gy, info.occupied)) {
                return FALSE;
            }

            calculatedPosition = [gx, gy];
            updateGrid(calculatedPosition, info); // Mark the spaces on the grid as filled

            return TRUE; // Return true so some() will stop and also return true.
        },
        /* putWord() processes each item on the list,
               calculate it's size and determine it's position, and actually
               put it on the canvas. */
        putWord = function(item, i) {
            var info = meta[i];
            item.layout = NULL;

            // not getting the info means we shouldn't be drawing this one.
            if (!info) {
                return FALSE;
            }

            var obj = getOccupiedArrayNBound(info);
            info.occupied = obj.occupied;
            var bound = (info.bound = obj.bound);

            // skip the loop if we have already know the bounding box of
            // word is larger than the canvas.
            if (bound[2] - bound[0] + 1 > ngx || bound[3] - bound[1] + 1 > ngy) {
                return FALSE;
            }

            // Determine the position to put the text by
            // start looking for the nearest points
            var r = maxRadius + 1;

            while (r--) {
                var points = getPointsAtRadius(maxRadius - r);

                // Try to fit the words by looking at each point.
                // array.some() will stop and return true
                // when putWordAtPoint() returns true.
                // If all the points returns false, array.some() returns false.
                var drawn = points.some(tryToPutWordAtPoint, info);

                if (drawn) {
                    var x = calculatedPosition[0],
                        y = calculatedPosition[1],
                        { bound } = info;

                    let polygon = [];
                    info.occupied.forEach(o => {
                        if (o[0] && o[1]) {
                            polygon.push([o[0] * g, o[1] * g]);
                        }
                    });

                    item.layout = {
                        x: (x + info.gw / 2) * g,
                        y: (y + info.gh / 2) * g,
                        width: info.width,
                        height: info.height,
                        rotation: info.rotation,
                        size: info.size,
                        nodeList: info.nodeList,
                        bb: info.bb,
                        tightBound: {
                            x: bound[0] * g,
                            y: bound[1] * g,
                            width: (bound[2] - bound[0] + 1) * g,
                            height: (bound[3] - bound[1] + 1) * g
                        },
                        polygon
                    };

                    updateBound(drawnBound, [x * g, y * g], [(x + info.gw) * g, (y + info.gh) * g]);
                    return TRUE;
                }
            }
            // we tried all distances but text won't fit, return false
            return FALSE;
        },
        precalculate = function() {
            var availablePlotArea = maskImageData ? areaOfMask : geometryUtils.areaOfEllipse(plotWidth / 2, plotHeight / 2),
                sum = 0,
                outWidth = 0,
                outHeight = 0,
                maxBoxWidth = 0,
                maxBoxHeight = 0,
                percent,
                fontArea = 0,
                range = fontScale.range(),
                domain = fontScale.domain(),
                maxFontSize = range[range.length - 1],
                _fonts = {};
            rotationCollection = data.map(_rotation);

            let writeFunc = (d, i) => {
                    var fontSize = _fontSize(d),
                        contentType = _content(d).type,
                        customFontSize = contentType === 'html' ? fontSize : 10;

                    _fonts[fontSize] = fontSize;
                    drawWordsInSvg(d, i, customFontSize);
                },
                readFunc = (d, i) => {
                    var fontSize = _fontSize(d),
                        padding = getWordPadding(d),
                        contentType = _content(d).type,
                        customFontSize = contentType === 'html' ? fontSize : 10,
                        fontRatio = fontSize / customFontSize,
                        deg = rotationCollection[i]; // ZC3305

                    var textEle = getDummyText(i),
                        bb = bbox(textEle.node()),
                        boxWidth = bb.width + padding,
                        boxHeight = bb.height + padding,
                        area = boxWidth * boxHeight,
                        scaledWidth = boxWidth * fontRatio * mathAbs(mathCos(deg)) + boxHeight * fontRatio * mathAbs(mathSin(deg)),
                        scaledHeight = boxWidth * fontRatio * mathAbs(mathSin(deg)) + boxHeight * fontRatio * mathAbs(mathCos(deg));

                    maxBoxWidth = mathMax(maxBoxWidth, scaledWidth);
                    maxBoxHeight = mathMax(maxBoxHeight, scaledHeight);
                    outHeight = mathMax(outHeight, scaledHeight - plotHeight);
                    outWidth = mathMax(outWidth, scaledWidth - plotWidth);
                    sum += area * mathPow(fontRatio, 2);
                    fontArea += (area * mathPow(fontRatio, 2)) / (fontSize / maxFontSize);
                };

            if (singleNode) {
                data.forEach((d, i) => {
                    writeFunc(d, i);
                    readFunc(d, i);
                });
            } else {
                data.forEach(writeFunc);
                data.forEach(readFunc);
            }

            if ((outWidth || outHeight) && sum < availablePlotArea) {
                var outWidthPercent = outWidth ? outWidth / plotWidth : 0,
                    outHeightPercent = outHeight ? outHeight / plotHeight : 0;

                percent = mathMax(outWidthPercent, outHeightPercent);
                percent = 1 - mathMin(percent, pick(maxSizeFactor, 0.2));
            } else {
                let factor = mathMin(
                    mathSQRT(availablePlotArea / sum),
                    plotWidth / maxBoxWidth,
                    plotHeight / maxBoxHeight,
                    mathSQRT(availablePlotArea / fontArea)
                ); // ZC2859, FEATURES_BAR_13
                percent = factor * (config.boundBox ? 0.5 : 1);
                percent = mathMax(percent, pick(maxSizeFactor, 0.4));
            }

            var minFontSize = 3,
                uniqueFontSizes = objectValues(_fonts),
                r0 = pick(
                    config.size.min,
                    mathMax(uniqueFontSizes.length === 1 && uniqueFontSizes[0] === range[0] ? range[0] * percent : range[0], minFontSize)
                ),
                r1 = pick(config.size.max, mathMax(range[range.length - 1] * percent, r0 + 2)),
                tempScalePoint = d3_scalePoint();

            tempScalePoint.domain(domain).range([r0, r1]);
            fontScale.range(domain.map(tempScalePoint));
        },
        updatPixelByColor = function(imageData, i, rgb) {
            if (rgb) {
                imageData[i] = rgb.r;
                imageData[i + 1] = rgb.g;
                imageData[i + 2] = rgb.b;
                imageData[i + 3] = mathMax(1, mathFloor(rgb.opacity * 255));
            }
        },
        operations = function() {
            imageDataX = imageDataY = canvasCount = imageTotalWidth = imageTotalHeight = currentRowHeight = 0;
            canvasMeta = [{ width: 1, height: 1 }];
            textImageData = [];
            drawnBound = [plotWidth / 2, plotHeight / 2, plotWidth / 2, plotHeight / 2];

            precalculate();
            createGrid();
            if (singleNode) {
                meta = data.map((d, i) => {
                    drawWordsInSvg(d, i);
                    return getTextInfo(d, i);
                });
                canvasMeta.forEach(getCanvas);
                data.forEach(drawWordsInCanvas);
                canvasMeta.forEach(readImageData);
                drawnData = data.filter(putWord);
                atEnd();
            } else {
                data.forEach((d, i) => drawWordsInSvg(d, i));
                breathNInvoke(() => {
                    meta = data.map(getTextInfo);
                    breathNInvoke(() => {
                        canvasMeta.forEach(getCanvas);
                        data.forEach(drawWordsInCanvas);
                        canvasMeta.forEach(readImageData);

                        breathNInvoke(() => {
                            drawnData = data.filter(putWord);
                            if (drawnData.length < data.length && adjust) {
                                breathNInvoke(drawAgain);
                            }
                            breathNInvoke(atEnd);
                        });
                    });
                });
            }
        },
        findScaleFactor = function() {
            if (config.boundBox) {
                let scaleFactor = mathMin(drawnBound[0] - 0, drawnBound[1] - 0, plotWidth - drawnBound[2], plotHeight - drawnBound[3]);
                return mathMin(plotWidth, plotHeight) / (mathMin(plotWidth, plotHeight) - scaleFactor);
            }

            let scale = d3_scaleLinear(),
                l = scale.domain([plotWidth / 2, drawnBound[0]])(0),
                t = scale.domain([plotHeight / 2, drawnBound[1]])(0),
                r = scale.domain([plotWidth / 2, drawnBound[2]])(plotWidth),
                b = scale.domain([plotHeight / 2, drawnBound[3]])(plotHeight);
            return mathMin(l, t, r, b);
        },
        drawAgain = function() {
            // ZC3032
            outside[1] = mathMax(0, (plotHeight - drawnBound[3]) / g);

            if (!outside[1]) {
                return;
            }
            drawnBound = [plotWidth / 2, plotHeight / 2, plotWidth / 2, plotHeight / 2];
            pointsAtRadius = [];
            grid = [];
            createGrid();
            drawnData = data.filter(putWord);
        },
        atEnd = function() {
            let scaleFactor = maskImageData || (defined(config.size.min) && defined(config.size.max)) ? 1 : findScaleFactor();
            config.onload({ mask: maskCanvas, scaleFactor });
            dummyText.remove();
            maskImageData = NULL;
        },
        draw = function() {
            data.sort(sortByFontSizeAccesor);
            if (adjust && data.length && _content(data[0]).type !== 'html') {
                data.sort((a, b) => (_fontSize(a) === _fontSize(b) ? sortByContentAccesor(a, b) : 1));
            }
            breathNInvoke(operations);
        },
        drawOnMask = function() {
            var imageData = maskContext.getImageData(0, 0, w, h),
                maskBackgroundColor = config.maskBackground,
                rgb = maskBackgroundColor && maskBackgroundColor !== 'inherit' && d3_rgb(maskBackgroundColor);
            maskImageData = imageData.data;
            areaOfMask = 0;

            for (var i = 0; i < maskImageData.length; i += 4) {
                if (maskImageData[i + 3]) {
                    areaOfMask++;
                    updatPixelByColor(maskImageData, i, rgb);
                }
            }

            if (rgb) {
                maskContext.putImageData(imageData, 0, 0);
            }

            draw();
        };

    // All set, start the drawing
    if (!mask) {
        draw();
    } else {
        var maskCanvas = createCanvas(),
            w = ngx * g,
            h = ngy * g;

        maskCanvas.width = w;
        maskCanvas.height = h;
        var maskContext = createContext(maskCanvas);

        let drawOnShape = function(_mask) {
            var size = mathMin(w, h),
                bound = _mask._name === 'rect' ? [w, h] : [size, size],
                path = _mask.apply(NULL, bound);

            maskContext.translate(w / 2, h / 2);
            maskContext.scale(w / bound[0], h / bound[1]);
            canvasUtils.drawPath(maskContext, path);

            maskContext.fillStyle = '#000';
            maskContext.fill();
            drawOnMask();
        };

        if (typeof mask === 'function') {
            drawOnShape(mask);
        } else {
            var img = new Image();
            img.crossOrigin = 'Anonymous'; // Uncaught DOMException: Failed to execute 'getImageData' on 'CanvasRenderingContext2D': The canvas has been tainted by cross-origin data.
            img.onload = function() {
                maskContext.drawImage(img, 0, 0, w, h);
                drawOnMask();
            };
            img.onerror = function() {
                maskCanvas = NULL;
                drawOnShape(getSVGsymbol());
            };

            img.src = mask;
        }
    }
};

Registry.setComponent('wordcloud.layout', WordCloud);
