// @zohocharts/zc/src/geo-heatmap/geo-heatmap.js
//heatmap-series.js
import { MapSeries } from '../geo-core/map-series';
import { inherit, pick, each } from '../utils/zcutils';
import { d3_set } from '../core/d3-constants';
import { GeoUtils } from '../geo-core/map-utils';
import { maps } from '../core/chart-instance-manager';
import { d3_values } from '../utils/zcutils';
import { labelRenderer } from '../core/main-zc-init';
import { ZoomPosition } from '../interfaces/zoom-position';
/**
 *
 * @param series
 * @param geoMap
 * @param index
 * @constructor
 */
export function GeoHeatMapSeries(series, geoMap, index) {
    this.renderingMode = geoMap.systemConf.map.renderingMode;
    MapSeries.call(this, series, geoMap, index);
}

// inherit from MapSeries
inherit(GeoHeatMapSeries, MapSeries);

// add HeatMapSeries to global holder
maps.GeoMap.SeriesTypes.set('geoheatmap', GeoHeatMapSeries);

/**
 *
 * @param reset
 */
GeoHeatMapSeries.prototype.render = function(reset) {
    var mapSeries = this.initialize(),
        geoMap = mapSeries.geoMap,
        NULL = null,
        mapFeatures = geoMap.dataObject.getUserMapFeatures(),
        plotOptions = mapSeries.plotOptions,
        gradient = plotOptions.gradients,
        defaultMapConfig = geoMap.systemConf.map,
        gradientEnabled = d3_set(['radial', 'linear']).has(gradient.type);

    mapSeries.features = [];
    mapSeries.seriesdataAfterUpdate = [];

    if (!mapSeries.series.disabled) {
        each(mapSeries.series.data, function(points, index) {
            points = geoMap.dataObject.updateSeriesdatabyCategories(d3_values(points));
            mapSeries.seriesdataAfterUpdate[index] = points;

            each(points, function(point, i) {
                var key = mapSeries.geoMap.dataset.getGeo(point);
                if (mapFeatures.has(key)) {
                    var currentMapFeature = mapFeatures.get(key),
                        gradientDimension = NULL;

                    if (gradientEnabled) {
                        var bbox_geo = GeoUtils.getFeatureGradientBBox(
                            currentMapFeature,
                            gradient.type,
                            mapSeries.geoMap.geo.utils.geoPath
                        );
                        gradientDimension = [bbox_geo.width, bbox_geo.height];
                        currentMapFeature.transform = {
                            x: bbox_geo.x,
                            y: bbox_geo.y
                        };
                    }

                    /**
                     * set style in each feature
                     *
                     * @type {{fillColor: *, strokeColor: *, strokeWidth: *, strokeStyle: *}}
                     */
                    currentMapFeature.styles = {
                        fillColor: mapSeries.fillColor(
                            mapSeries.index,
                            point,
                            NULL,
                            NULL,
                            mapSeries.plotOptions.fillOpacity,
                            gradientDimension
                        ),
                        strokeColor: mapSeries.strokeColor(
                            mapSeries.index,
                            point,
                            pick(plotOptions.strokeColor, defaultMapConfig.strokeColor),
                            NULL,
                            plotOptions.strokeOpacity,
                            i
                        ),
                        strokeWidth: plotOptions.strokeWidth,
                        strokeStyle: plotOptions.strokeStyle,
                        fillOpacity: pick(plotOptions.fillOpacity, 1),
                        strokeOpacity: pick(plotOptions.strokeOpacity, 1)
                    };

                    mapSeries.features.push(currentMapFeature);
                }
            });
        });
    }

    if (reset) {
        mapSeries.mapLayer.update(mapSeries.features);
    } else {
        mapSeries.mapLayer = mapSeries.geoMap.addLayer(
            mapSeries.features,
            NULL,
            `series-${mapSeries.index}`,
            'series',
            defaultMapConfig.useMeshOutline
        );
    }

    if (labelRenderer) {
        geoMap.extraParams.labelQueue.push({
            renderer: mapSeries,
            charttype: 'scatter',
            chartObj: geoMap,
            args: [mapSeries.series, mapSeries.index]
        });
    }

    geoMap.timerObj[geoMap.timerObj.length] = mapSeries.renderedTimer = setTimeout(function() {
        mapSeries.chartRendered = true;
    }, 17);
};

GeoHeatMapSeries.prototype.commonRendererProp = function() {
    return this.geoMap.rendererConf[this.index];
};

/**
 *
 * @param point
 * @return {ZoomPosition}
 */
GeoHeatMapSeries.prototype.getZoomPosition = function(point) {
    let geoMap = this.geoMap,
        transform = geoMap.behaviouralEvents.eventManager.behaviours.PLOT_ZOOM.instance.getTransform();

    if (geoMap.geo.isGlobeMap) {
        return new ZoomPosition('rotate', geoMap.dataObject.getLatLonPoint(point));
    } else {
        return new ZoomPosition(
            'bound',
            geoMap.geo.utils.geoPath.bounds(geoMap.dataObject.getFeatureByPoint(point)).map(b0 => transform.invert(b0))
        );
    }
};
