// @zohocharts/zc/src/behaviours/plot-zoom/plot-zoom-arguments.js
import { Registry } from '../../core/zc-registry';
import { mathMax } from '../../core/zc-constants';
import { getAxisObject } from '../../axis/axisutils';
import { MINIMUM_VISIBLE_POINTS } from '../behavioural-constants';
import { getMinimumVisibleArea } from './behaviour-constrainer';
import { d3_event } from '../../core/d3-constants';
import { $Browser } from '../../core/browser-detect';
import { pick } from '../../utils/zcutils';

const DEFAULT_MAX_ZOOM = 25;

/**
 *
 * @param chart
 * @return {ZoomD3Options}
 */
export function _getPlotZoomArguments(chart) {
    let options = chart.systemConf.chart.zoom,
        plotarea = chart.plotarea,
        translateExtent = [
            [0, 0],
            [plotarea.width, plotarea.height]
        ],
        ZoomD3Options = Registry.getComponent('ZoomD3Options'),
        zoomType = chart.isGeoMap ||  chart.dataObject.isFlowChartCategory ? 'xy' : options.type,
        maxZoom = options.maxZoom || getMaxZoom(chart, pick(options.minimumVisiblePoints, MINIMUM_VISIBLE_POINTS), zoomType);

    return new ZoomD3Options({
        type: zoomType,
        scaleExtent: [options.minZoom, maxZoom],
        translateExtent: translateExtent,
        extent: translateExtent,
        wheelZoom: options.wheelZoom !== false,
        pan: options.pan !== false,
        transitionDuration: options.transitionDuration,
        element: chart.eventHandler.trackerGroup,
        metaKey: options.metaKey,
        filter: function(event) {
            // #ZC2560
            const d3event = d3_event();
            if (d3event?.constructor?.name === 'BrushEvent') {
                return false;
            }

            if ($Browser.isTouchDevice) {
                return event.constructor.name === 'WheelEvent' || event.touches.length > 1 || event.changedTouches.length > 1;
            }
            d3_event(event);
            return true;
        }
    });
}

/**
 *
 * @param chart
 * @param numberOfPointsInView
 * @param zoomType
 */
function getMaxZoom(chart, numberOfPointsInView = 3, zoomType = 'xy') {
    if (zoomType === 'xy') {
        return DEFAULT_MAX_ZOOM;
    }

    let scale = getAxisObject(chart, zoomType, 0).scale,
        side = chart.axes.rotated ? 'height' : 'width';

    if (zoomType === 'x' || (zoomType === 'y' && scale.bandwidth)) {
        return mathMax(1, chart.plotarea[side] / getMinimumVisibleArea(chart, zoomType, numberOfPointsInView));
    }

    return DEFAULT_MAX_ZOOM;
}

Registry.setComponent('_getPlotZoomArguments', _getPlotZoomArguments);
