// @zohocharts/zc/src/line/canvasLinechart.js
//$Id$
//canvasLinechart.js
import { utils, validateNullVal, mergeJSON, defined, pick, d3_values } from '../utils/zcutils';
import { getStrokeColor } from '../utils/colorutils';
import { gradient_helpers } from '../effects/gradient-helpers';
import { getDashArray, getLineCap } from '../utils/dashstyles';
import { datacount_based_intelligence, dimension_based_intelligence } from '../utils/intelligence-utils';
import { getLineLayout } from './line-layout';
import { mathMin } from '../core/zc-constants';
import { NULL } from '../core/zc-constants';
import { ORDINAL_DATATYPE } from '../core/datatype-constants';
import { charttype, canvas_plot_renderer } from '../core/main-zc-init';
import { LineHighlighter } from './lineHighlight';
import { DataProcessor } from '../core/data-processor';
import { getChartTypeByIndex } from '../core/charttype-helpers';
import { isFacet } from '../axis/hierarchy-axis-utils';

canvas_plot_renderer.line = function() {
    'use strict';
    var commonRendererProp,
        shiftInfo,
        yrangePos,
        scatter = canvas_plot_renderer.scatter?.();

    function chart(context, series, pattern, filterIndex) {
        var rp = commonRendererProp.params,
            chartObj = rp.chartObj,
            xscale = rp.xscale,
            yscale = rp.yscale,
            index = rp.datasetIndex,
            legendOrder = rp.legendOrderIndex,
            rendererIndex = rp.rendererIndex,
            axisrotated = rp.axisrotated,
            commonPlotOptions = rp.commonPlotOptions,
            color = rp.color,
            labelQueue = rp.labelQueue,
            yaxiscolumnorder = rp.yaxiscolumnorder,
            // isSelectionContainer = !(chartObj.container.select('.selection-canvas').empty()),
            isSelectionCanvas = context.canvas.className == 'selection-canvas' ? true : false,
            selectionArray = chartObj.cache.clickSelectionData;

        var currentdata = series;
        currentdata = chartObj.dataObject.categoryHierarchy ? series[index] : chartObj.seriesdata[index];
        var plotoptions = currentdata.plotoptions;
        var gvp = (commonRendererProp.params.gvp = utils.getValueOfPlotOptions(plotoptions, commonPlotOptions)),
            xReversed = chartObj.systemConf.chart.axes.xaxis.reversed,
            gradientDim = gradient_helpers.getDimension(chartObj, index),
            canvasarea = chartObj.canvasarea;
        var paneid = isFacet(chartObj) ? `${currentdata.row}_${currentdata.column}` : '0_0';
        var tempChart = getChartTypeByIndex(chartObj, paneid, index),
            chartTypeNumeric = tempChart.value,
            isAreaChart = chartTypeNumeric != charttype.line;
        var dashStyle = gvp('dashStyle'),
            strokeWidth = validateNullVal(gvp('strokeWidth'), 2),
            strokeOpacity = gvp('strokeOpacity'),
            strokeColor = getStrokeColor(chartObj, legendOrder, paneid, gradientDim, rp),
            markeroptions = (commonRendererProp.params.markerOptions = plotoptions
                ? mergeJSON(commonPlotOptions.marker, plotoptions.marker, true)
                : commonPlotOptions.marker),
            strokeStyle,
            hoverOptions = (commonRendererProp.params.hoverOptions =
                plotoptions && plotoptions.hoveroptions
                    ? mergeJSON(commonPlotOptions.hoveroptions, plotoptions.hoveroptions, true)
                    : commonPlotOptions.hoveroptions);

        var strokeLineCap = getLineCap(gvp('lineCap'), dashStyle),
            strokeDashArray = getDashArray(dashStyle, strokeWidth),
            markerEnabled =
                datacount_based_intelligence(chartObj, 'marker') &&
                dimension_based_intelligence(chartObj, 'marker', mathMin(canvasarea.width, canvasarea.height))
                    ? markeroptions.enabled && scatter  // don't call scatter when the scatter renderer is not available
                    : false;

        var datatypeX = chartObj.dataset.getDataType('x'),
            datatypeY = chartObj.dataset.getDataType('y', yaxiscolumnorder),
            xCategories = datatypeX == ORDINAL_DATATYPE ? xscale.domain() : null,
            yCategories = datatypeY == ORDINAL_DATATYPE ? yscale.domain() : null;

        if (defined(pattern)) {
            context.lineWidth = pick(pattern['stroke-width'], strokeWidth);

            if (defined(filterIndex)) {
                markerEnabled = true;
                strokeOpacity = 0.00000001;
            }
        } else {
            context.lineWidth = strokeWidth;
        }
        if (!currentdata.disabled) {
            chart.seriesdataAfterUpdate = [];
            context.save();
            currentdata.data.forEach(function(d, i) {
                context.beginPath();
                var serData;
                if (datatypeX == ORDINAL_DATATYPE || datatypeY == ORDINAL_DATATYPE) {
                    serData = chartObj.dataObject.updateSeriesdatabyCategories(d3_values(d), xCategories, yCategories, yaxiscolumnorder);
                } else {
                    serData = d;
                }
                if (chartObj.renderVisiblePointsOnly) {
                    serData = DataProcessor.helpers.filterPointsInViewport(chartObj, serData, 'line', 'x', index);
                }

                chart.seriesdataAfterUpdate[i] = serData;
                getLineLayout(rp, serData, shiftInfo, yrangePos, paneid).context(context)(serData);

                if (strokeDashArray.length && typeof strokeDashArray === 'object' && defined(context.setLineDash)) {
                    context.setLineDash(strokeDashArray);
                    context.lineCap = strokeLineCap;
                }
                context.strokeStyle = strokeColor(legendOrder, NULL, gvp('strokeColor') || color, NULL, strokeOpacity, i);

                if(isSelectionCanvas && defined(selectionArray[0]) && chartObj.dataObject.totalnoofdatas != chartObj.cache.selection.selectedPoints.length){
                    context.strokeStyle = strokeColor(legendOrder, NULL, pattern, NULL, strokeOpacity, i);
                }

                context.stroke();
            });
            context.restore();

            if (markerEnabled) {
                scatter.commonRendererProp(commonRendererProp).shiftInfo(shiftInfo).yrangePos(yrangePos);
                scatter.chartRendered = false;
                scatter(context, series, pattern, filterIndex);
            }
        }
        var markerSize = 0;
        if (!markerEnabled) {
            markerSize = markerEnabled ? (markeroptions.size || strokeWidth) * 0.5 * 10 + strokeWidth / 2 : 0;
            if (chartObj.dataObject.markerSize[tempChart.name]) {
                chartObj.dataObject.markerSize[tempChart.name].add(markerSize);
            }

            labelQueue.push({
                renderer: chart,
                charttype: 'scatter',
                chartObj: chartObj,
                rendererIndex: rendererIndex,
                paneid,
                args: [currentdata, index, markerSize]
            });
        }

        chart.getSymbolSize =
            scatter && scatter.getSymbolSize ||
            function() {
                return markerSize;
            };
        chart.chartRendered = true;
        context = null;     // memory leak fix
        return chart;
    }
    chart.getHighlighter = function(chartObj, index) {
        if (!chart.highlighter) {
            chart.highlighter = new LineHighlighter(chartObj, chart);
        }
        return chart.highlighter;
    };
    chart.commonRendererProp = function(_) {
        return arguments.length ? (commonRendererProp = _, chart) : commonRendererProp;
    };
    chart.shiftInfo = function(_) {
        return arguments.length ? (shiftInfo = _, chart) : shiftInfo;
    };
    chart.yrangePos = function(_) {
        return arguments.length ? (yrangePos = _, chart) : yrangePos;
    }
    return chart;
};

export var _canvasLineChart = canvas_plot_renderer.line;
