// @zohocharts/zc/src/line/lineOptions.js
import { defaultTheme as options } from '../core/defaulttheme';

options.line = function() {
    return {
        mode: 'linear', //linear,stepBefore,stepAfter,cardinal
        strokeWidth: 2,
        outerPadding: 0.02,
        stacked: {
            enabled: false,
            showAs: 'values' //values,percent
        },
        marker: {
            enabled: true,
            innerStrokeWidth: 2,
            outerFillColor: 'transparent',
            outerStrokeColor: 'transparent',
            useSVGMarkers: true
        },
        hoveroptions: {
            innerStrokeWidth: 2,
            outerFillOpacity: 0.5,
            outerStrokeOpacity: 0.1
        },
        datalabels: {
            showAs: 'y' //x,y,percent
        },
        animation: {
            type: 'strokeTween', //strokeTween,horiPathTween,vertPathTween,pointbypoint,pointTween
            marquee: {
                enabled: false,
                duration: 300
            },
            blink: {
                enabled: false,
                type: 'twinkle', //twinkle | fade
                duration: 800
            }
        },
        gradients: {
            // linear || none
            options: {
                linear: {
                    x2: 10,
                    spreadMethod: 'reflect' // pad || repeat || reflect
                }
            }
        }
    };
};

export var _lineOptions = options.line;
