// @zohocharts/zc/src/line/linechart.js
//$Id$
//linechart.js
import { svg_plot_renderer, charttype, chartcategory } from '../core/main-zc-init';
import { DataProcessor } from '../core/data-processor';
import { utils, validateNullVal, mergeJSON, generateIdByPropValue, d3_values, defined, simpleClone, invokeFunction } from '../utils/zcutils';
import { gradient_helpers } from '../effects/gradient-helpers';
import { colorscale_helpers } from '../axis/colorscale-helpers';
import { getAnimationConf, interTween } from '../utils/animationutils';
import { datacount_based_intelligence, dimension_based_intelligence } from '../utils/intelligence-utils';
import { $Browser } from '../core/browser-detect';
import { ORDINAL_DATATYPE, TIME_DATATYPE } from '../core/datatype-constants';
import { getLineLayout } from './line-layout';
import { appendEle } from '../utils/zcsvgutils';
import { addSVGmarker, blinkSVGsymbols } from '../utils/svg-symbols';
import {
    STROKETWEEN_ANIMATION,
    POINTBYPOINT_ANIMATION,
    POINTTWEEN_ANIMATION,
    HORIPATHTWEEN_ANIMATION,
    VERTPATHTWEEN_ANIMATION,
    VERTPATHTWEENALL_ANIMATION
} from '../core/animation-constants';
import { effects } from '../effects/zceffects';
import { NULL, NONE, mathMin, mathMax } from '../core/zc-constants';
import { getStrokeColor, getRawColor } from '../utils/colorutils';
import { getDashArray, getLineCap, applyMarquee } from '../utils/dashstyles';
import { d3_interpolate, d3_select } from '../core/d3-constants';
import { START_ALIGN, MIDDLE_ALIGN, END_ALIGN } from '../core/align-constants';
import { LineHighlighter } from './lineHighlight';
import { getChartTypeByIndex } from '../core/charttype-helpers';
import { setOpacityForTickFilterElements } from '../plot/svg/set-pivot-attr';
import { isFacet } from '../axis/hierarchy-axis-utils';

/**
 *
 * @type {chart} line renderer.
 */
svg_plot_renderer.line = function() {
    var commonRendererProp, xscale, yscale, index, dataset, axisrotated, yaxiscolumnorder, color;
    var line = [],
        scatter = svg_plot_renderer.scatter && svg_plot_renderer.scatter();
    var series,
        redraw = false,
        path,
        strokeDashArray,
        marqueeEnabled,
        marqueeDuration,
        marker;
    var stacked, labelQueue, commonPlotOptions, systemconf, hasNegValue, chartObj;
    var oldData, shiftInfo, yrangePos;
    var transitionDuration, animationType;

    function chart(selection) {
        selection.each(function(data) {
            //renderer config values
            var rp = commonRendererProp.params;
            (chartObj = rp.chartObj),
                (xscale = rp.xscale),
                (yscale = rp.yscale),
                (stacked = rp.stacked),
                (index = rp.datasetIndex),
                (axisrotated = rp.axisrotated),
                (commonPlotOptions = rp.commonPlotOptions),
                (color = rp.color),
                (labelQueue = rp.labelQueue),
                (hasNegValue = rp.hasNegValue),
                (yaxiscolumnorder = rp.yaxiscolumnorder),
                (dataset = rp.dataset),
                (systemconf = chartObj.systemConf);

            var renderingOrder = rp.renderingOrderIndex,
                legendOrder = rp.legendOrderIndex,
                rendererIndex = rp.rendererIndex,
                { shapescale } = rp;

            var defsLocation = chartObj.defsLocation;

            if(chartObj.customVisual){
                chart.chartRendered = true;
                return;
            }
            var parser_data, data1;     // line-marker parser case
            if (chartObj.viewFlow && chartObj.viewFlow.marks._default.chart.length) {
                parser_data = data[index];
                data1 = parser_data.data;
            }
            var currentdata = data1 || data[index];
            var data_bind = { data: currentdata, index: renderingOrder };
            stacked? data_bind.stackIndex = ((data.length * 1) + renderingOrder) : null;
            var plotoptions = currentdata.plotoptions;
            var currentseriesdata = currentdata.data;
            var paneid = isFacet(chartObj) ? `${currentdata.row}_${currentdata.column}` : '0_0';
            var tempChart = getChartTypeByIndex(chartObj, paneid, index),
                cType = tempChart.name,
                chartTypeNumeric = tempChart.value,
                dataProcessorHelpers = DataProcessor.helpers;
            var gvp = (commonRendererProp.params.gvp = utils.getValueOfPlotOptions(plotoptions, commonPlotOptions)),
                isAreaChart = chartTypeNumeric != charttype.line,
                isRangeChart = cType == 'arearange',
                isPolar = chartObj.dataObject.isPolarAxisCategory,
                gradientDim = gradient_helpers.getDimension(chartObj, index),
                canvasarea = chartObj.canvasarea;
            var interpolation = gvp('mode'),
                dashStyle = gvp('dashStyle'),
                strokeWidth = validateNullVal(gvp('strokeWidth'), 2),
                strokeOpacity = gvp('strokeOpacity'),
                strokeColor = getStrokeColor(chartObj, index, paneid, gradientDim, rp),
                hasColorScale = colorscale_helpers.isQuantileColorScale(chartObj),
                markeroptions = (commonRendererProp.params.markerOptions =
                    plotoptions && plotoptions.marker
                        ? mergeJSON(commonPlotOptions.marker, plotoptions.marker, true)
                        : commonPlotOptions.marker),
                hoverOptions = (commonRendererProp.params.hoverOptions =
                    plotoptions && plotoptions.hoveroptions
                        ? mergeJSON(commonPlotOptions.hoveroptions, plotoptions.hoveroptions, true)
                        : commonPlotOptions.hoveroptions);

            //Animation variables
            var animationOptions = getAnimationConf(commonPlotOptions, plotoptions, chartObj, cType);
            var isAnimationEnable = animationOptions.enabled,
                easingType = animationOptions.easingType;
            transitionDuration = animationOptions.duration;
            animationType = animationOptions.type;
            animationType = animationType == VERTPATHTWEENALL_ANIMATION ? VERTPATHTWEEN_ANIMATION : animationType;

            var blinkConf =
                plotoptions && plotoptions.animation && plotoptions.animation.blink
                    ? mergeJSON(commonPlotOptions.animation.blink, plotoptions.animation.blink, true)
                    : commonPlotOptions.animation.blink;
            var race_syncTrans = chartObj.racing && chartObj.racing.syncTransition,
                keyFramedRace = chartObj.racing && chartObj.racing.keyFramedRace;
            var markerEnabled =
                    hasColorScale ||
                    (datacount_based_intelligence(chartObj, 'marker') &&
                    dimension_based_intelligence(chartObj, 'marker', mathMin(canvasarea.width, canvasarea.height))
                        ? markeroptions.enabled
                        : false),
                strokeLineCap = getLineCap(gvp('lineCap'), dashStyle),
                callScatterRenderer =
                    $Browser.IE ||
                    $Browser.SAFARI ||
                    interpolation == 'stepBefore' ||
                    interpolation == 'stepAfter' ||
                    interpolation == 'step-before' ||
                    interpolation == 'step-after' ||
                    rp.thresholdType ||
                    hasColorScale ||
                    shapescale ||
                    ($Browser.FIREFOX && blinkConf.enabled) ||
                    (markeroptions.useSVGMarkers === false && scatter); // IE & SAFARI don't support marker updation. So, render the scatter chart instead of using marker.; // don't call scatter when the scatter renderer is not available
            strokeDashArray = getDashArray(dashStyle, strokeWidth);

            var marqueeOptions =
                plotoptions && plotoptions.animation && plotoptions.animation.marquee
                    ? mergeJSON(commonPlotOptions.animation.marquee, plotoptions.animation.marquee, true)
                    : commonPlotOptions.animation.marquee;
            marqueeEnabled = marqueeOptions.enabled;
            marqueeDuration = marqueeOptions.duration;

            var datatypeX = chartObj.dataset.getDataType('x'),
                datatypeY = chartObj.dataset.getDataType('y', yaxiscolumnorder),
                xCategories = datatypeX == ORDINAL_DATATYPE ? xscale.domain() : null,
                yCategories = datatypeY == ORDINAL_DATATYPE ? yscale.domain() : null,
                categories =
                    datatypeX == ORDINAL_DATATYPE && !defined(chartObj.dataObject.hierarchical)
                        ? xCategories
                        : chartObj.dataObject.getCategories('x');

            var yReversed = systemconf.chart.axes.yaxis[yaxiscolumnorder].reversed;
            if (datatypeX != ORDINAL_DATATYPE && categories.length == 0 && isAnimationEnable && animationType == POINTTWEEN_ANIMATION) {
                categories = chartObj.dataObject.findXCategories();
            }
            var prevData = chart.seriesdataAfterUpdate ? chart.seriesdataAfterUpdate[0]: null;
            var prevSeriesDataInfo = chartObj.racing ? simpleClone(chart.seriesdataAfterUpdate) : null;
            chart.seriesdataAfterUpdate = [];
            currentseriesdata.forEach(function(d, i) {
                var serData;
                if (datatypeX == ORDINAL_DATATYPE || datatypeY == ORDINAL_DATATYPE) {
                    serData = chartObj.dataObject.updateSeriesdatabyCategories(d3_values(d), xCategories, yCategories, yaxiscolumnorder);
                } else {
                    serData = d;
                }

                chart.seriesdataAfterUpdate[i] = serData;
                line[i] = getLineLayout(rp, serData, shiftInfo, yrangePos, paneid);
            });

            var container = d3_select(this);
            chart.update = function() {
                container.transition().call(chart);
            };
            var dynamicPosition = rp.dynamicSeriesPosition;
            if (dynamicPosition === undefined) {
                dynamicPosition = index;
            }
            let insertBefore;
            if (stacked) {
                // For ordering series to fix overlapping of markers
                let lineseries_n = this.querySelector(`g.lineseries_${dynamicPosition-1}`);
                insertBefore = lineseries_n ? `g.${lineseries_n.className.baseVal} + *` : null;
            }
            series = appendEle(container, 'g', [data_bind], '.', 'class', `lineseries_${dynamicPosition}`, undefined, undefined, {insertBefore : insertBefore});
            series.node().index = index;
            if (isRangeChart) {
                series = appendEle(series, 'g', [data_bind], '.', 'class', `rangePos_${yrangePos}`);
            }

            var markerID,
                mSize = isPolar
                    ? {
                          inner: 0.32,
                          outer: 0.56
                      }
                    : {
                          inner: 0.4,
                          outer: 0.7
                      };
            if (markerEnabled && !callScatterRenderer) {
                markeroptions.color = currentdata.color || getRawColor(chartObj, legendOrder, currentdata, index, paneid, rendererIndex);
                var updatedMarkerOptions= colorCustomFunctionInvoker(markeroptions);
                function colorCustomFunctionInvoker(options){
                    var clonedOptions=simpleClone(options);
                    clonedOptions.outerStrokeColor=invokeFunction(clonedOptions.outerStrokeColor,options.color,clonedOptions,chartObj);
                    clonedOptions.innerStrokeColor=invokeFunction(clonedOptions.innerStrokeColor,options.color,clonedOptions,chartObj);
                    clonedOptions.outerFillColor=invokeFunction(clonedOptions.outerFillColor,options.color,clonedOptions,chartObj);
                    clonedOptions.innerFillColor=invokeFunction(clonedOptions.innerFillColor,options.color,clonedOptions,chartObj);
                    return clonedOptions;
                }
                markerID = generateIdByPropValue(markeroptions, 'marker') + generateIdByPropValue(blinkConf, 'blink');
                marker = addSVGmarker(
                    series,
                    markerID,
                    updatedMarkerOptions,
                    {
                        color: updatedMarkerOptions.color,
                        symbolSize: mSize,
                        strokeWidth: strokeWidth
                    },
                    chartObj
                );
                delete markeroptions.color;
            }

            if (markerEnabled && !callScatterRenderer) {
                blinkSVGsymbols(marker.selectAll('.outer,.inner'), blinkConf, chartObj);
            }

            if (markerEnabled && callScatterRenderer) {
                scatter
                    .commonRendererProp(commonRendererProp)
                    .shiftInfo(shiftInfo)
                    .yrangePos(yrangePos);
                scatter.chartRendered = false;

                container.call(scatter);
            } else {
                container.selectAll(`g.scatterseries_${dynamicPosition}`).remove();
            }
            container = null;   // removing container reference for memory leak issue

            // console.log(chartObj.axes.x.scale.domain());
            currentseriesdata.forEach(function(d, i) {
                var ishideable = currentdata.disabled;

                // render the visible points only
                if (chartObj.renderVisiblePointsOnly) {
                    chart.seriesdataAfterUpdate[i] = DataProcessor.helpers.filterPointsInViewport(
                        chartObj,
                        chart.seriesdataAfterUpdate[i],
                        'line',
                        'x',
                        index
                    );
                }
                
                var serData = chart.seriesdataAfterUpdate[i];
                if (chartObj.systemConf.chart.axes.xaxis.reversed) {
                    serData = serData.slice(0).reverse();
                }
                var baseData = serData;

                var basedataShift,
                    dummyDataLength = 0;

                var subseries = appendEle(
                    series,
                    'g',
                    function(d) {
                        return ishideable ? [] : [d];
                    },
                    '.',
                    'class',
                    `subseries_${i}`
                );

                path = subseries.selectAll('path.line').data(function(d) {
                    return ishideable ? [] : [d];
                });
                if (isAnimationEnable && animationType != STROKETWEEN_ANIMATION) {
                    if (animationType == POINTTWEEN_ANIMATION) {
                        //get baseData for starting animation
                        baseData =
                            oldData != null
                                ? oldData.map(function(d, j) {
                                      return Object.create(oldData[0]);
                                  })
                                : serData.map(function(d, i) {
                                      return Object.create(serData[0]);
                                  });
                    } else {
                        var idx = animationType == HORIPATHTWEEN_ANIMATION ? dataset.xidx : dataset.yidx(yaxiscolumnorder);
                        var scaleObj = animationType == HORIPATHTWEEN_ANIMATION ? xscale : yscale;
                        var baseIdx = isPolar && yReversed && !stacked ? scaleObj.domain().length - 1 : 0;
                        baseData = serData.map(function(d, i) {
                            var s = Object.create(d);
                            if (d.isDummy) {
                                dummyDataLength++;
                            }
                            if (stacked && animationType != HORIPATHTWEEN_ANIMATION) {
                                s.y0 = 0;
                                s.y =
                                    !isPolar && hasNegValue != null && scaleObj(hasNegValue) != null
                                        ? mathMax(hasNegValue, scaleObj.domain()[0])
                                        : scaleObj.domain()[0];
                            } else {
                                if (isRangeChart && animationType == VERTPATHTWEEN_ANIMATION) {
                                    var yArr = new Array(2);
                                    var y0 = chartObj.dataset.getY(d, yaxiscolumnorder, 0);
                                    yArr[0] = y0;
                                    yArr[1] =
                                        s[idx] == null
                                            ? null
                                            : hasNegValue != null && scaleObj(hasNegValue) != null
                                            ? mathMax(hasNegValue, y0)
                                            : y0;
                                    s[idx] = yArr;
                                } else {
                                    s[idx] =
                                        s[idx] == null
                                            ? null
                                            : !isPolar && hasNegValue != null && scaleObj(hasNegValue) != null
                                            ? mathMax(hasNegValue, scaleObj.domain()[0])
                                            : scaleObj.domain()[baseIdx];
                                }
                            }
                            return s;
                        });
                    }
                    basedataShift = d3_values(baseData);
                }

                if (shiftInfo != null) {
                    if (redraw) {
                        var clipX = xscale(serData[0][0]) + (xscale.bandwidth != null ? xscale.bandwidth() / 2 : 0);
                        var clipWid =
                            chartObj.dataObject.getXYvalue(
                                axisrotated,
                                stacked,
                                xscale,
                                yscale,
                                yaxiscolumnorder,
                                serData[serData.length - 1],
                                'x',
                                null,
                                yrangePos
                            ) -
                            chartObj.dataObject.getXYvalue(
                                axisrotated,
                                stacked,
                                xscale,
                                yscale,
                                yaxiscolumnorder,
                                serData[0],
                                'x',
                                null,
                                yrangePos
                            );

                        var properties = {
                                name: 'rect',
                                id: chartObj.id,
                                x: clipX,
                                y: 0,
                                width: clipWid,
                                height: chartObj.plotarea.height
                            },
                            args = {
                                chartObj: chartObj
                            };
                        var clipID = effects.get(effects.svg_clips, 'rect', properties, args);
                        series.attr('clip-path', `url(${defsLocation}#${clipID})`);
                    }
                }

                var pathEnter = path
                    .enter()
                    .append('path')
                    .attr('class', 'line')
                    .datum(baseData, function(d) {
                        return dataset.getX(d);
                    })
                    .attr('d', line[i])
                    .style('stroke', currentdata.color || color)
                    .style('fill', NONE)
                    .style('stroke-width', strokeWidth)
                    .style('stroke-opacity', strokeOpacity == null || strokeOpacity === '' ? 1 : strokeOpacity);

                if (isAnimationEnable) {
                    if (animationType == POINTBYPOINT_ANIMATION || animationType == POINTTWEEN_ANIMATION) {
                        interTween(
                            baseData,
                            basedataShift,
                            serData,
                            categories,
                            animationType,
                            easingType,
                            transitionDuration,
                            rp,
                            function(d) {
                                return {
                                    d: line[i](d)
                                };
                            },
                            pathEnter
                        );
                    } else if (animationType == STROKETWEEN_ANIMATION) {
                        strokeTween(pathEnter, easingType, transitionDuration);
                    } else {
                        let isEmpty = !serData.length || (serData.length == 1 && !serData[0].length);
                        isEmpty = stacked ? serData.length == dummyDataLength : isEmpty;
                        if (isEmpty) {
                            pathEnter.attr('d', function(d) {
                                return line[i]([]);
                            });
                        } else {
                            if(race_syncTrans){
                                pathEnter
                                    .transition(race_syncTrans)
                                    .attrTween('d', pathTween);
                            }else{
                                pathEnter
                                    .transition()
                                    .duration(transitionDuration)
                                    .ease(easingType)
                                    .attrTween('d', pathTween);
                            }
                        }
                    }
                }

                var pathUpdate = pathEnter
                    .merge(path)
                    .datum(serData, function(d) {
                        return dataset.getX(d);
                    })
                    .attrs({
                        id: `line_${index}`,
                        'stroke-dasharray': strokeDashArray
                    })
                    .styles({
                        'stroke-linecap': strokeLineCap,
                        'stroke-width': strokeWidth,
                        'stroke-opacity': validateNullVal(strokeOpacity, 1),
                        fill: NONE,
                        stroke: strokeColor(index, NULL, gvp('strokeColor') || color, NULL, strokeOpacity, i)
                    });

                var dataUpdate = serData;

                if (shiftInfo != null) {
                    var transX =
                        shiftInfo != null
                            ? Math.abs(
                                  chartObj.dataObject.getXYvalue(
                                      axisrotated,
                                      stacked,
                                      chart.oldXscale,
                                      yscale,
                                      yaxiscolumnorder,
                                      serData[0],
                                      'x',
                                      null,
                                      yrangePos
                                  ) -
                                      chartObj.dataObject.getXYvalue(
                                          axisrotated,
                                          stacked,
                                          xscale,
                                          yscale,
                                          yaxiscolumnorder,
                                          serData[0],
                                          'x',
                                          null,
                                          yrangePos
                                      )
                              )
                            : null;
                    path.datum(dataUpdate)
                        .attr('d', line[i])
                        .transition()
                        .duration(transitionDuration)
                        .ease(easingType)
                        .attr('transform', function() {
                            return transX == null ? null : `translate(-${transX},0)`;
                        });

                    resetDashProp(path);
                } else if (pathEnter.node() == null) {
                    if (race_syncTrans && !keyFramedRace && path.node()) {                        
                        let { maxRange } = chartObj.systemConf.chart.axes.xaxis,
                            xType = chartObj.dataset.getDataType('x'),
                            isEntirePlotRace =
                                (xType === 'time' &&
                                    chartObj.scroll.brushScale.bandwidth &&
                                    !chartObj.racing.isEntireAxisNeededForOrdinal) ||
                                (xType === 'ordinal' && !chartObj.racing.isEntireAxisNeededForOrdinal) ||
                                ((xType === 'time' || xType === 'numeric') && !maxRange);

                        var isCurvedLine = (commonPlotOptions.mode === 'cardinal' || commonPlotOptions.mode === 'monotone'),
                            isLinearLine = commonPlotOptions.mode === 'linear'
                        if (isCurvedLine || isLinearLine) {
                            if (chartObj.racing.interrupt) {
                                path.interrupt();
                            }
                            var toPath;
                            path.datum(dataUpdate)
                                .attr('d', function(d, i) {
                                    var from = this.getAttribute('d'),
                                        _f = from;

                                    toPath = line[i](d, i);

                                    if (from === null || toPath === null) {
                                        return toPath;
                                    }
                                    var isCatXorYisNull = point => {
                                            var dataProcessorHelpers = DataProcessor.helpers,
                                                dataset = chartObj.dataset,
                                                isCatNullX = dataProcessorHelpers.isCatNull(dataset, 'x'),
                                                isCatNullY = dataProcessorHelpers.isCatNull(dataset, 'y', yaxiscolumnorder);
                                            if (point) {
                                                return isCatNullX(dataset.getX(point)) || isCatNullY(dataset.getY(point, yaxiscolumnorder));
                                            }
                                        },
                                        prevData = prevSeriesDataInfo[i],
                                        prevDataLength = prevData.length,
                                        prevEndPoint = this.getPointAtLength(this.getTotalLength()),
                                        minLen_toCurve = 3,
                                        isMinLen_toCurve = (dataUpdate.length === minLen_toCurve) && (prevData.length < minLen_toCurve),
                                        isMinLen_toCurve_afterNull =
                                            isCatXorYisNull(prevData[prevDataLength - minLen_toCurve]) &&
                                            !isCatXorYisNull(prevData[prevDataLength - 2]) &&
                                            !isCatXorYisNull(prevData[prevDataLength - 1]),
                                        endPoint = prevEndPoint.x + ',' + prevEndPoint.y,
                                        appendPointsToFromPath = (from, endPoint, prevEndPoint) => {
                                            for (var _i = 0; _i < d.length - prevSeriesDataInfo[i].length; _i++) {
                                                if (isCurvedLine) {
                                                    from = from + 'C ' + endPoint + ',' + endPoint + ',' + endPoint;
                                                } else {
                                                    from = from + 'L ' + prevEndPoint.x + ' ' + prevEndPoint.y;
                                                }
                                            }
                                            return from;
                                        };

                                    if ((isMinLen_toCurve || isMinLen_toCurve_afterNull) && isCurvedLine) {
                                        from = from + ',' + endPoint + ',' + endPoint;
                                    }
                                    from = appendPointsToFromPath(from, endPoint, prevEndPoint);

                                    if (
                                        prevSeriesDataInfo[i].length > 2 &&
                                        isCatXorYisNull(prevData[prevDataLength - (minLen_toCurve - 1)])
                                    ) {
                                        this.setAttribute('d', from + 'Z');
                                        prevEndPoint = this.getPointAtLength(this.getTotalLength());
                                        endPoint = prevEndPoint.x + ',' + prevEndPoint.y;
                                        from = appendPointsToFromPath(_f, endPoint, prevEndPoint);
                                    }
                                    return from;
                                })
                                .attr('transform', null)
                                .transition(race_syncTrans)
                                .ease(easingType)
                                .attr('d', toPath);
                            resetDashProp(path);
                            
                        } else {
                            var prev_len = path.node().getTotalLength();
                            if (chartObj.racing.interrupt) {
                                path.interrupt();
                            }
                            path.datum(dataUpdate)
                                .attr('transform', null)
                                .attr('d', line[i]);

                            path.styles({
                                'stroke-dashoffset': function(d, i) {
                                    if (dataUpdate.length < prevSeriesDataInfo[i].length) {
                                        return this.getTotalLength() || -1;
                                    }
                                    return this.getTotalLength() - prev_len || -1;
                                },
                                'stroke-dasharray': function() {
                                    return `${this.getTotalLength()} ${this.getTotalLength()}`;
                                }
                            });
                            path.transition(race_syncTrans)
                                .style('stroke-dashoffset', function() {
                                    return "0";
                                })
                                .on('end', function() {
                                    resetDashProp(d3_select(this));
                                });
                        }
                    } else if(race_syncTrans){
                        path.datum(dataUpdate)
                        .attr('transform', null)
                        .transition(race_syncTrans)
                        .attr('d', line[i]);

                        /*
                        * Reset stroke-dasharray & stroke-dashoffset
                        * If animation.type = "strokeTween" & calling redraw before the animation completes
                        */
                        resetDashProp(path);
                    } else {
                        path.datum(dataUpdate)
                            .attr('transform', null)
                            .transition()
                            // .delay(((isAreaChart && ((animationType != POINTTWEEN_ANIMATION) || (animationType == POINTTWEEN_ANIMATION && redraw != false && currentdata.disabled != false))) ? 0 : transitionDuration))
                            .duration(transitionDuration)
                            .ease(easingType)
                            .attr('d', line[i]);

                        /*
                         * Reset stroke-dasharray & stroke-dashoffset
                         * If animation.type = "strokeTween" & calling redraw before the animation completes
                         */
                        resetDashProp(path);
                    }
                }
                var mID = markerEnabled && !callScatterRenderer ? `url(${defsLocation}#${markerID})` : NULL,
                    markerPosition = markeroptions.position,
                    markerProperties = markerPosition
                        ? {
                              'marker-start': markerPosition.indexOf(START_ALIGN) > -1 ? mID : NULL,
                              'marker-mid': markerPosition.indexOf(MIDDLE_ALIGN) > -1 ? mID : NULL,
                              'marker-end': markerPosition.indexOf(END_ALIGN) > -1 ? mID : NULL
                          }
                        : {
                              marker: mID
                          };

                pathUpdate.styles(markerProperties);
                setOpacityForTickFilterElements(pathUpdate, chartObj, 'line');
                if (chartObj.dataObject.hierarchical) {
                    pathUpdate.attr('levelbycolumn', function(d, i) {
                        return d[0] && d[0].levelByColumn;
                    });
                    if (chartObj.dataObject.pivot) {
                        pathUpdate.attr('levelbyrow', function(d, i) {
                            return d[0] && d[0].levelByRow;
                        });
                    }
                }

                pathEnter.exit().remove();

                function pathTween() {
                    var interpolate = d3_interpolate(line[i](baseData), line[i](serData));
                    return function(t) {
                        return interpolate(t);
                    };
                }
            });

            var markerSize = 0;
            if (!(callScatterRenderer && markerEnabled)) {
                markerSize = markerEnabled ? (markeroptions.size || strokeWidth) * 0.5 * 10 + strokeWidth / 2 : 0; //common 0.5 sizefactor for scatter, line, threshold
                labelQueue.push({
                    renderer: chart,
                    charttype: 'scatter',
                    chartObj: chartObj,
                    rendererIndex: rendererIndex,
                    paneid: paneid,
                    args: [currentdata, index, markerSize, yrangePos]
                });
                //Initiated in dataprocessor
                if (chartObj.dataObject.markerSize[cType]) {
                    chartObj.dataObject.markerSize[cType].add(markerSize);
                }
            }

            chart.getSymbolSize =
                scatter && scatter.getSymbolSize ||
                function() {
                    return markerSize;
                };
        });
        let chartInstance = commonRendererProp.params.chartObj;
        chartInstance.timerObj[chartInstance.timerObj.length] = chart.renderedTimer = setTimeout(
            function() {
                if (marqueeEnabled) {
                    applyMarquee(series.selectAll(`#line_${index}`), marqueeDuration, strokeDashArray, chart);
                }
                chart.chartRendered = true;
                redraw = true;
            },
            !redraw ? transitionDuration + 20 : 2 * transitionDuration + 20
        );

        chart.oldXscale = xscale;
        return chart;
    }

    //chart.prototype = new $ZC.renderer();
    function strokeTween(path, easingType, duration) {
        path.styles({
            'stroke-dasharray': function() {
                return `${this.getTotalLength()} ${this.getTotalLength()}`;
            },
            'stroke-dashoffset': function() {
                return this.getTotalLength();
            }
        })
            .transition()
            .duration(duration)
            .ease(easingType)
            .style('stroke-dashoffset', "0")
            .on('end', function() {
                resetDashProp(d3_select(this));
            });
    }
    //stroke-dashoffset is considered to be string, sincein firefox the tweening must happen between values with type matched, either string ot string, or numeric to numeric

    function resetDashProp(path) {
        path.styles({
            'stroke-dasharray': strokeDashArray,
            'stroke-dashoffset': null
        });
    }
    chart.getHighlighter = function(chartObj, index) {
        if (!chart.highlighter) {
            chart.highlighter = new LineHighlighter(chartObj, chart);
        }
        return chart.highlighter;
    };

    chart.commonRendererProp = function(_) {
        if (!arguments.length) {
            return commonRendererProp;
        }
        commonRendererProp = _;
        return chart;
    };

    chart.shiftInfo = function(_) {
        if (!arguments.length) {
            return shiftInfo;
        }
        shiftInfo = _;
        return chart;
    };

    chart.yrangePos = function(_) {
        if (!arguments.length) {
            return yrangePos;
        }
        yrangePos = _;
        return chart;
    };

    chart.destroy = function() {
        series = path = labelQueue = null;
        scatter && scatter.destroy();
    }

    return chart;
};

export var _lineChart = svg_plot_renderer.line;
