// @zohocharts/zc/src/bar/bar-selection.js
import { pointInRectangle, distanceBetweenRectFromPoint } from '../events/events-utils';
import { d3_ascending } from '../core/d3-constants';
import { NULL } from '../core/zc-constants';
import { getRotatedMouse } from '../events/eventActions';

export function barSelection(points, mouse, chart, key) {
    var point;
    var commonPlotOptions = chart ? chart.systemConf.chart.plot.plotoptions[key] : null,
        customTriggerOffset = commonPlotOptions && commonPlotOptions.events && commonPlotOptions.events.triggerOffset,
        sizeOffset = customTriggerOffset || 0,
        _needRect2PtDist = sizeOffset && needRect2PtDistCalc(chart),
        _needExtendBound = sizeOffset && supportedExtendedBound(chart);
    var x = points.map((p, i) => {
        let _a = p._data(),
            bound_default = getBound(_a),
            rect = _needExtendBound ? getExtendedBound(bound_default, sizeOffset) : bound_default,
            _mouse = getRotatedMouse(mouse, { x: _a._x, y: _a._y, rotation: _a._rotation });

        var _b = pointInRectangle(_mouse.x, _mouse.y, rect),
            distance;
        if (_needRect2PtDist && _b) {
            /*
              distance between bar and mouse point
             */
            p.searchKey = 'bar_triggeroffset';
            distance = distanceBetweenRectFromPoint(bound_default, _mouse);
        }
        return [_b ? 0 : 1, _a, distance];
    });
    var nearbtPts = x.filter(d => {
        if (d[0] == 0) {
            return d;
        }
    });

    if (nearbtPts.length > 1) {
        var high,
            rank = {
                marker: 0,
                //for normal data, rank is 1
                levelMarker: 2
            };
        if (chart && chart.dataObject.isHierarchy){
            var levelOrder = 0, nearest;
            nearbtPts.map(function(d){
                var _ord = d[1].level.match(/\d+/g).length;
                if(_ord > levelOrder){
                    levelOrder = _ord;
                    nearest = [d];
                }
            })
            nearbtPts = nearest;
        }
        var sorted = nearbtPts.sort(function(a, b) {
            if (_needRect2PtDist) {
                /*
                sort by distance calculated above
                For multi clustered bar chart, to avoid overlapping of bounds,
                we need to calculate by minimum distance between bar and mouse
                 */
                return d3_ascending(a[2], b[2]);
            } else {
                var rank_p0 = rank[a[1].type] || 1,
                    rank_p1 = rank[b[1].type] || 1;
                return rank_p0 == rank_p1 ? -1 : d3_ascending(rank_p0, rank_p1);
            }
        });
        point = sorted[0][1];
        return point;
    } else if (nearbtPts.length) {
        point = nearbtPts[0][1];
        return point;
    }

    return NULL;
}

function getBound(point) {
    let { _x: x, _y: y, _width: width, _height: height, _centric: centric } = point;

    if (centric) {
        return { x: x - width / 2, y: y - height / 2, width, height };
    }
    return { x, y, width, height };
}

function getExtendedBound(rect, extendedDist) {
    var x = (rect.x || 0) - extendedDist,
        y = (rect.y || 0) - extendedDist,
        width = rect.width + 2 * extendedDist,
        height = rect.height + 2 * extendedDist;

    return { x, y, width, height };
}

function needRect2PtDistCalc(chart) {
    if (chart) {
        return (
            chart.dataObject.noofBarEnabledSeries &&
            chart.dataObject.noofBarEnabledSeries > 1 &&
            (chart.chartTypes[0].name != 'bullet' && chart.chartTypes[0].name != 'heatmap')
        );
    }
    return null;
}

function supportedExtendedBound(chart) {
    if (chart) {
        return chart.dataObject.noofBarEnabledSeries && (chart.chartTypes[0].name != 'bullet' && chart.chartTypes[0].name != 'heatmap');
    }
    return null;
}
export function _barPointSelection(point, mouse) {
    let rect = getBound(point),
        _mouse = getRotatedMouse(mouse, { x: point._x, y: point._y, rotation: point._rotation }),
        _b = pointInRectangle(_mouse.x, _mouse.y, rect);
    return _b;
}
