// @zohocharts/zc/src/bar/bar-highlight.js
import { RECT_ELEMENT } from '../core/element-constants';
import { barSelection } from './bar-selection';
import { _super, defined, getPropVal, pick } from '../utils/zcutils';
import { charttype } from '../core/main-zc-init';
import { appendEle, roundedRect } from '../utils/zcsvgutils';
import { getRawColor, isMultiColoring } from '../utils/colorutils';
import { patternFillEffect } from '../events/visualEffects/highlight/pattern-fill';
import { HIGHLIGHT_CONTEXT } from '../events/visualEffects/highlight/highlight-context';
import { getCorrectedRectParams } from '../utils/zcsvgutils';
import { isPointVisible } from '../events/events-utils';

export function BarHighlighter(chart, renderer) {
    this.chart = chart;
    this.renderer = renderer;
}

var proto = BarHighlighter.prototype;

proto._defaultH = function() {
    return [
        {
            shape: RECT_ELEMENT,
            path: this.path,
            data: this.data,
            context: this.renderingContext,
            translate: this.translateContext,
            styles: patternFillEffect
        }
    ];
};

proto.data = function(point) {
    var chart = this.chart,
        renderer = this.renderer,
        rp = renderer.commonRendererProp().params;
    return function() {
        if (!chart.cache.data[point._key]) {
            chart.cache.data[point._key] = {};
        }
        var cache = chart.cache.data[point._key],
            transPoint;
        if (!defined(cache._x)) {
            let x = chart.dataObject.getTransRectValue('x', point, point.subSeriesIndex, 0, renderer.commonOption),
                y = chart.dataObject.getTransRectValue('y', point, point.subSeriesIndex, 0, renderer.commonOption),
                w = chart.dataObject.getTransRectValue('width', point, point.subSeriesIndex, 0, renderer.commonOption),
                h = chart.dataObject.getTransRectValue('height', point, point.subSeriesIndex, 0, renderer.commonOption);

            const parser = renderer.commonRendererProp().parser,
                parserCoordinates = parser && parser.helpers.positionComponent[parser.KEY].coordinates;
            if (parserCoordinates) {
                // The coordinates of the bar is updated from view parser and saved in point.updatedCoordinates
                let coord = parserCoordinates(point, rp.datasetIndex, {x, y, width: w, height: h});
                coord = getCorrectedRectParams(coord.x, coord.y, coord.width, coord.height, null, rp.yaxisreversed, rp.axisrotated)
                point.updatedCoordinates = {x: coord.x1, y: coord.y1, widthT: coord.x2, heightT: coord.y2, width: coord.width, height: coord.height};
            } else {
                delete point.updatedCoordinates;
            }
            
            transPoint = getCorrectedRectParams(x, y, w, h, null, rp.yaxisreversed, rp.axisrotated);

            cache._x = point._x = cache._x || transPoint.x1;
            cache._y = point._y = cache._y || transPoint.y1;
            cache._width = point._width = cache._width || w;
            cache._height = point._height = cache._height || h;
            cache._widthT = point._widthT = cache._widthT || transPoint.x2; //Translated screen x2, y2 value needed for roundedRect Function, but absolute width, height needed for barSelection. In future need to change roundedrect x2, y2
            cache._heightT = point._heightT = cache._heightT || transPoint.y2;
            cache._uid = point._uid = chart.dataset.getX(point);
            cache._point = point;
        }
        cache._visible = point._visible = isPointVisible(chart, point);
        return point;
    };
};

proto.options = function(point) {
    var chart = this.chart,
        renderer = this.renderer,
        rp = renderer.commonRendererProp().params,
        border = rp.gvp('border') || {},
        radius = border.show ? border.radius : 0,
        color = getRawColor(
            this.chart,
            isMultiColoring(rp.commonPlotOptions, this.chart) ? pick(point.categoryIndex, point.sortedItemIndex) : point.seriesIndex,
            point,
            point.seriesIndex
        );
    const parser = renderer.commonRendererProp().parser,
        parserRadius = parser && parser.helpers.positionComponent[parser.KEY].radius;
    radius = parserRadius ? parserRadius(point, rp.datasetIndex, radius) : radius;
    return {
        radius: chart.dataObject.getTransRectValue('radius', null, null, null, renderer.commonOption, radius),
        color: color
    };
};

proto.renderingContext = function() {
    return HIGHLIGHT_CONTEXT.get('bar').call(this, this.chart);
};

proto.translateContext = function() {
    return [this.chart.plotarea.left, this.chart.plotarea.top];
};

proto.path = function(point, model) {
    var yaxisReversed = this.chart.systemConf.chart.axes.yaxis[point.yaxisOrder || 0].reversed;
    point = point._data();
    let x = point._x, y = point._y, widthT = point._widthT, heightT = point._heightT;
    point.updatedCoordinates ? ({x, y, widthT, heightT} = point.updatedCoordinates): null;
            
    var rect = roundedRect(
        x,
        y,
        widthT,
        heightT,
        this.options(point).radius,
        yaxisReversed,
        this.chart.axes.rotated,
        true
    );
    var context = model.context.call(this),
        pathContext = appendEle(context, 'path', [1], '.', 'class', `highlight${point.paneid}${point.seriesIndex}${point.subSeriesIndex}${point.itemIndex}`);
    context.style('display', '');
    pathContext
        .style('display', '')
        .attr('d', rect)
        .attr('transform', `translate(${this.chart.plotarea.left},${this.chart.plotarea.top})`);
    return pathContext;
};
