// @zohocharts/zc/src/bar/bar-encode.js
import { defined, getBorderRadius, mergeJSON, pick } from '../utils/zcutils';
import { _setOpacityForTickFilterElements } from '../plot/svg/set-pivot-attr';
import { getFillColor, getStrokeColor, isMultiColoring } from '../utils/colorutils';
import { gradient_helpers } from '../effects/gradient-helpers';
import { mathMin } from '../core/zc-constants';
import { d3_select } from '../core/d3-constants';
import { checkRadiusWithinBoundary } from '../utils/zcsvgutils';

export let bar_position = function() {

    let position_helpers = {};

    const KEY = this.KEY || 'bar';
    this.KEY = KEY;

    let borderComponent = this.builderC.border;

    this.helpers.position[KEY] = {};

    // constant dependencies
    const axesRotated = this.chart.axes.rotated,
        yaxisreversed = this.builderC.yaxisreversed;
    /*
    START POSITION
     */
    this.helpers.position[KEY].start = (d, encode) => {
        let { stacked, hasNegValue, yaxiscolumnorder } = this.builderC;
        let chartPhase = this.chart.phase;
        let x, y, width, height;
        let radius = borderComponent && borderComponent.show ? borderComponent.radius : null;
        radius = this.helpers.positionComponent[KEY].radius(d, this.builderC.datasetIndex, radius);

        let x_data = this.chart.dataset.getX(d),
            yRange_low = this.scales.y.range()[0],
            y_negValue = hasNegValue != null ? this.scales.y(hasNegValue) : null,
            rangeY_0 = this.rangeYType ? this.scales.y(this.chart.dataset.getY(d, 0, 0)) : null,
            // rangeY_0 = this.rangeYType ? this.scales.y(this.chart.dataset.getY(d, yaxiscolumnorder, 0)) : null,
            y_priority = pick(rangeY_0, y_negValue);

        x = stacked
            ? this.scales.x(x_data) + this.scales.interX(this.commonOption.stackGroupIndex) + this.commonOption.padding
            : this.scales.x(x_data) + this.scales.interX(this.barDatasetIndex + (d.subseriesIdx || 0)) + this.commonOption.padding;
        if (this.scales.y.bandwidth) {
            y = this.scales.y.origRange[0];
            // y = this.rangeYType ? rangeY_0 + this.scales.y.bandwidth() / 2 : this.scales.y.origRange[0];
        } else {
            if (y_priority != null) {
                y = y_priority;
            } else {
                if (stacked && chartPhase == 'legendFilter') {
                    yRange_low = this.scales.y(d.y0);
                }
                let y_zeroValue = this.scales.y(0);
                y = this.chart.axes.rotated || stacked ? yRange_low : y_zeroValue != null ? mathMin(y_zeroValue, yRange_low) : yRange_low;
            }
        }

        width = this.chart.axes.rotated ? 0 : this.commonOption.barWidth;
        height = this.chart.axes.rotated ? this.commonOption.barWidth : 0;

        ({x, y, width, height} = this.helpers.positionComponent[KEY].coordinates(d, this.builderC.datasetIndex, {x, y, width, height}));
        let rect = this.chart.axes.rotated ? { x: y, y: x, width, height, radius } : { x, y, width, height, radius };

        if (isNaN(x)) {
            return getAbsRectPartialCoords(rect, this.builderC.yaxisreversed, this.chart.axes.rotated, this.chart.axes.rotated ? 'y' : 'x');
        }
        if (isNaN(y)) {
            return getAbsRectPartialCoords(rect, this.builderC.yaxisreversed, this.chart.axes.rotated, this.chart.axes.rotated ? 'x' : 'y');
        }
        if (!encode) {
            let coords = getAbsRectCoords(rect, this.builderC.yaxisreversed, this.chart.axes.rotated);
            return coords;
        }
    };

    /*

    END POSITION

     */
    this.helpers.position[KEY].end = (d, encode, yRangePos, relative) => {
        let rectValue = this.chart.dataObject.getTransRectValue;
        let radius = borderComponent && borderComponent.show ? borderComponent.radius : null;
        radius = this.helpers.positionComponent[KEY].radius(d, this.builderC.datasetIndex, radius);

        if (encode) {
            if (encode == 'radius') {
                return radius;
            }
            return this.chart.dataObject.getTransRectValue(encode, d, d.subseriesIdx || 0, null, this.commonOption, null, yRangePos);
        }
        let x, y, width, height;

        (x = this.chart.dataObject.getTransRectValue('x', d, d.subseriesIdx || 0, null, this.commonOption)),
            (y = this.chart.dataObject.getTransRectValue('y', d, d.subseriesIdx || 0, null, this.commonOption, yRangePos)),
            (width = this.chart.dataObject.getTransRectValue('width', d, d.subseriesIdx || 0, null, this.commonOption)),
            (height = this.chart.dataObject.getTransRectValue('height', d, d.subseriesIdx || 0, null, this.commonOption));

        let yColInfo = this.chart.dataset.getColumn('y', this.builderC.yaxiscolumnorder);
        let isLogScale = yColInfo.scaleType === 'log';
        if (isLogScale) {
            var range = this.scales.y.range();
            var _start = this.scales.y.invert(range[0]),
                _end = this.scales.y.invert(range[range.length - 1]);
            height = _start === _end ? 0 : height;
        }
        ({x, y, width, height} = this.helpers.positionComponent[KEY].coordinates(d, this.builderC.datasetIndex, {x, y, width, height}));
        if (relative) {
            return { x, y, width, height };
        }
        let coords = getAbsRectCoords({ x, y, width, height, radius }, this.builderC.yaxisreversed, this.chart.axes.rotated);
        d.oldCoordinates = coords;
        return coords;
    };

    /*

    INTERMEDIATE STOP 1 - CHAIN ANIMATED

     */
    this.helpers.position[KEY].stop_1 = (d, encode, yRangePos) => {
        let start = this.helpers.position[KEY].start(d),
            end = this.helpers.position[KEY].end(d);
        let x1, y1, x2, y2;
        if (this.chart.axes.rotated) {
            y1 = end.y1;
            y2 = end.y2;
            return { y1, y2 };
        } else {
            x1 = end.x1;
            x2 = end.x2;
            return { x1, x2 };
        }
        // let coords = getAbsRectCoords({ x, y, width, height }, this.builderC.yaxisreversed, this.chart.axes.rotated);
        return { x1, y1, x2, y2 };
    };

    this.helpers.position[KEY].exit = (d, encode) => {
        const coords = { ...d.oldCoordinates };
        axesRotated ? (yaxisreversed ? (coords.x1 = coords.x2) : (coords.x2 = coords.x1))
            : (yaxisreversed ? (coords.y2 = coords.y1) : (coords.y1 = coords.y2));

        return coords;
    };

    return position_helpers;
};

/*

ENCODE FOR RACING BAR


 */
export function racing_bar_position() {
    let position_helpers = {};

    const KEY = 'bar_racing',
        SUPER_KEY = 'bar';

    let borderComponent = this.builderC.border,
        radius = borderComponent && borderComponent.show ? borderComponent.radius : null;

    this.helpers.position[KEY] = {};

    this.helpers.position[KEY].start = (d, encode, yRangePos) => {
        var yOrient = this.chart.axes.y.groups[0].prerenderComp.orient;
        if (this.chart.axes.rotated && yOrient === 'bottom') {
            let coords = this.helpers.position[SUPER_KEY].end(d);
            coords.y1 = -this.chart.plotarea.top;
            coords.y2 = this.scales.x.bandwidth() - this.chart.plotarea.top;
            return coords;
        }
        let coords = this.helpers.position[SUPER_KEY].end(d);
        coords.y1 = this.chart.plotarea.height + 20;
        coords.y2 = this.scales.x.bandwidth() + this.chart.plotarea.height + 20;
        return coords;
    };

    this.helpers.position[KEY].end = (d, encode, yRangePos) => {
        return this.helpers.position[SUPER_KEY].end(d);
    };

    this.helpers.position[KEY].exit = (d, encode, yRangePos) => {
        var yOrient = this.chart.axes.y.groups[0].prerenderComp.orient;
        if (this.chart.axes.rotated && yOrient === 'bottom') {
            let y1 = -this.chart.plotarea.top;
            let y2 = this.scales.x.bandwidth() - this.chart.plotarea.top;
            return { y1, y2 };
        }
        let y1 = this.chart.plotarea.height + 20;
        let y2 = this.scales.x.bandwidth() + this.chart.plotarea.height + 20;
        return { y1, y2 };
    };
}

export function getAbsRectCoords(rect, yReverse, rotated) {
    let r = rect.radius ? checkRadiusWithinBoundary(getBorderRadius(rect.radius), rect.width, rect.height) : 0;
    return {
        x1: yReverse && rotated ? rect.x - rect.width : rect.x,
        y1: yReverse && !rotated ? rect.y - rect.height : rect.y, //x1,y1 should be at top left . x2,y2 should be at bottom right .
        x2: yReverse && rotated ? rect.x : rect.x + rect.width,
        y2: yReverse && !rotated ? rect.y : rect.y + rect.height,
        radius: r
    };
}
export function getAbsRectPartialCoords(rect, yReverse, rotated, type) {
    let r = rect.radius ? checkRadiusWithinBoundary(getBorderRadius(rect.radius), rect.width, rect.height) : 0;
    if (type == 'x') {
        return {
            y1: yReverse && !rotated ? rect.y - rect.height : rect.y, //x1,y1 should be at top left . x2,y2 should be at bottom right .
            y2: yReverse && !rotated ? rect.y : rect.y + rect.height
        };
    } else {
        return {
            x1: yReverse && rotated ? rect.x - rect.width : rect.x,
            x2: yReverse && rotated ? rect.x : rect.x + rect.width
        };
    }
}
export let bar_styles = function() {
    const KEY = this.KEY || 'bar';
    const self = this;
    let chart = this.chart,
        renderer = this.renderer ?? this.oldRenderer,
        commonOptions = this.config.common,
        fillComponent = getBarFillColor.call(this), //TODO move this function here. avoid calling .call(this)
        strokeComponent = getStrokeColor(this.chart, this.builderC.datasetIndex, this.paneid, this.builderC),
        borderComponent = this.builderC.border || {};
    let isPolar = chart.dataObject.isPolarAxisCategory,
        multiColoring = isPolar ? commonOptions.multiColoring : isMultiColoring(commonOptions, this.chart),
        patternPallete = this.chart.patternPallete && this.chart.patternPallete.length;

    renderer.fillColorFunc = fillComponent; //FEATURESEVENTS_TICKLABEL_6

    this.helpers.styleComponent[KEY] = {
        fillComponent
    };
    this.helpers.positionComponent = {};
    this.helpers.positionComponent[KEY] = {
        radius: (d, i, defaultRadius) => defaultRadius,
        coordinates: (d, i, defaultCoordinates = {x, y, width, height}) => defaultCoordinates
    }
    this.helpers.styles[KEY] = {
        fill: (d, i) => {
            let itemPos = this.chart.systemConf.chart.plot.allowDuplicateCategories
                ? this.builderC.xcategoriesReversed.indexOf(this.chart.dataset.getX(d))
                : pick(d.sortedItemIndex, i);
            let baseClrIdx = multiColoring ? itemPos : this.legendOrder;
            let dim = patternPallete ? this.helpers.position.bar.end(d, null, null, true) : null;
            return fillComponent(baseClrIdx, d, null, this.builderC.paneColor, null, null, dim);
        },
        stroke: (d, i) => {
            // if (borderComponent.show) {
            let itemPos = this.chart.systemConf.chart.plot.allowDuplicateCategories
                ? this.builderC.xcategoriesReversed.indexOf(this.chart.dataset.getX(d))
                : pick(d.sortedItemIndex, i);
            let baseClrIdx = multiColoring ? itemPos : self.legendOrder;
            return strokeComponent(baseClrIdx, d, borderComponent.color || this.builderC.paneColor);
            // }
        },
        strokeWidth: borderComponent.show ? pick(borderComponent.size, 1) : 0,
        strokeDasharray: borderComponent.show ? borderComponent.style : null,
        fillOpacity: this.builderC.gvp('fillOpacity'),
        opacity: function(d, i) {
            let opacity = _setOpacityForTickFilterElements(d, i, chart, 'bar');
            d3_select(this).attr('filtered-opacity', opacity);
            return opacity;
        }
    };
};

function getBarFillColor() {
    let gradientDim = gradient_helpers.getDimension(this.chart, this.builderC.datasetIndex, this.paneid);
    return getFillColor(this.chart, this.builderC.datasetIndex, this.paneid, gradientDim);
}

function getBarBorder() {
    let seriesOptions = this.config.series,
        commonOptions = this.config.common;
    this.builderC.border =
        seriesOptions && seriesOptions.border ? mergeJSON(commonOptions.border, seriesOptions.border, true) : commonOptions.border;
    return this.builderC.border;
}
