// @zohocharts/zc/src/bar/bar-marks-mapping.js
import { createElement } from '../utils/zcsvgutils';
// import { getStackGroupIndex } from '../core/data-processor/transforms/stack/stack-grouping';
import { stack } from '../core/main-zc-init';
import { ParserRegistry } from '../core/parser-registry';

export let ZC_BAR_MARKS_MAPPING = function() {
    let _this = this;
    let mapping = {
        bar: {
            operator: ParserRegistry.marks('bar.default')
        },
        subdivider: {
            operator: ParserRegistry.marks('bar.subdivider')
        },
        levelmarker: {
            condition: () => {
                return _this.builderC.levelMarker && _this.builderC.levelMarker._zc_enabled;
            },
            operator: ParserRegistry.marks('bullet.levelmarker'),
            holder: () => {
                if (_this.builderC.stacked) {
                    let stackIndex = stack.getStackGroupIndex(_this.chart.dataObject.getStackGroup(), _this.builderC.datasetIndex);
                    return createElement('g', `.levelMarkerSeries_${stackIndex}`, [this.chart.seriesdata], _this.container, ':first-child');
                }
            }
        },
        targetmarker: {
            condition: () => _this.builderC.targetMarker && _this.builderC.targetMarker.enabled,
            operator: ParserRegistry.marks('bullet.targetmarker'),
            holder: () => _this.chart.staticPaneSVG.selectAll(`g.qualitativeGroup`)
        },
        whisker: {
            condition: () => _this.builderC.whiskers && _this.builderC.whiskers.enabled,
            operator: ParserRegistry.marks('boxplot.whisker'),
            holder: () => _this.chart.staticPaneSVG.selectAll(`g.qualitativeGroup`)
        },
        median: {
            condition: () => _this.builderC.median && _this.builderC.median.enabled,
            operator: ParserRegistry.marks('boxplot.median'),
            holder: () => _this.chart.staticPaneSVG.selectAll(`g.qualitativeGroup`)
        },
        radialbar: {
            operator: ParserRegistry.marks('radialbar.default')
        },
        marker: {
            condition: () => _this.builderC.commonPlotOptions.marker?.enabled,
            operator: ParserRegistry.marks('gantt.marker'),
            holder: () => _this.chart.staticPaneSVG.selectAll(`g.qualitativeGroup`)
        },
        connector: {
            condition: () => {
                return _this.builderC.commonPlotOptions.dependency?.enabled;
            },
            operator: ParserRegistry.marks('gantt.connector'),
        }
    };
    return mapping;
};
