// @zohocharts/zc/src/bar/bar-morph-collector.js
import { zcMorph } from '../morphing/zc-morping';
import { getOffsetValue } from '../plot/morph/morphTools';
import { d3_line } from '../core/d3-constants';
import { roundedRect } from '../utils/zcsvgutils';
import { MarksRegistry } from '../plot/MarksRegistry';
import { getBorderRadius } from '../utils/zcutils';

export function bar_morph_collectors() {
    let _this = this,
        { builderC, chart } = this,
        morphConfig = chart.systemConf.chart.plot.morph,
        morph_Obj = new zcMorph({ sampling: morphConfig.samplingDistance });

    let oldData = chart.oldSeriesInfo.get(builderC.datasetIndex),
        oldInfo = {
            seriesIdx: oldData.seriesIndex,
            color: oldData.color,
            transform: oldData.transformationInfo
        },
        offsetValue = getOffsetValue(chart, oldInfo.seriesIdx),
        transitionDuration = morphConfig.duration;

    var interpolated = {};
    var line = d3_line();

    let current_data = this.helpers.filterData(this.data.data[0]); //subseriesIdx
    current_data.map(function(d, i) {
        var params = _this.helpers.position.bar.end(d);
        if (isNaN(params.x1) || isNaN(params.y1)) {
            params.x1 = -1;
            return null;
        }
        let _radius = builderC.border && builderC.border.show ? builderC.border.radius : 0;

        var currentPath = roundedRect(params.x1, params.y1, params.x2, params.y2, getBorderRadius(_radius), null, null, true);
        var fromPath = chart.oldSeriesInfo.get(oldInfo.seriesIdx).pathData[i];
        var fromPathAttribute = fromPath.getAttribute('d');
        fromPath.setAttribute('d', currentPath);
        var toPath = fromPath.cloneNode();
        fromPath.setAttribute('d', fromPathAttribute);
        interpolated[i] = morph_Obj.interpolate(
            {
                path: fromPath,
                xOffset: offsetValue.x + oldInfo.transform[i].x,
                yOffset: offsetValue.y + oldInfo.transform[i].y
            },
            { path: toPath, xOffset: 0, yOffset: 0 }
        );
    });
    _this.helpers.position.bar_morphing = {
        interpolator: interpolated,
        line_d3: line
    };
    return interpolated;
}

MarksRegistry.add('bar.morph', bar_morph_collectors);
