// @zohocharts/zc/src/bar/bar-parser.js
import { getBarCommonOption } from './barRendererConf';
import { renderDom } from '../mark/dom-render';
import { bar_parser_helpers } from './bar-parser-helpers';
import { setBarBuilderConfig } from './bar-builder-config';
import { pick } from '../utils/zcutils';
import { createElement } from '../utils/zcsvgutils';
import { executeMarkOperators } from '../plot/mark-operator-executor';
import { ZC_BAR_MARKS_MAPPING } from './bar-marks-mapping';
import { bar_encode_collectors } from './bar-encode-helpers';
import { BARRANGE_CHARTNAME, GANTT_CHARTNAME } from '../core/chartname-constants';

export class BarRenderer {
    constructor(args) {
        let { data, config, rendererIndex, legendOrder, paneid, chart, builderC, barDatasetIndex, phase, type } = args;
        this.data = data;
        this.config = config;
        this.paneid = paneid;
        this.rendererIndex = rendererIndex;
        this.barDatasetIndex = barDatasetIndex;
        this.conf = chart.rendererConf[rendererIndex];
        this.legendOrder = legendOrder;
        this.builderC = builderC;
        this.chart = chart;
        this.type = type;
        this.phase = phase;
        this.oldRenderer = args.oldRenderer;
        this.markOperator = args.operator;
        this.setMarkMapping();
    }

    initialize(container) {
        this.rangeYType = (this.type == BARRANGE_CHARTNAME || this.type == GANTT_CHARTNAME) ? true : false;
        this.container = container;
        this.commonOption = getBarCommonOption(this.conf, this.rendererIndex, this.legendOrder, this.paneid);
        setBarBuilderConfig.call(this);
        this.seriesdataAfterUpdate = new Array(0);
        this.oldRenderer.seriesdataAfterUpdate = new Array(0);
        this.scales = {
            x: this.commonOption.bandwidth1,
            interX: this.commonOption.bandwidth2,
            y: this.commonOption.yscale
        };
        this.helpers = bar_parser_helpers.call(this);
        this.collectEncode();
    }
    collectEncode() {
        bar_encode_collectors.call(this);
    }
    getAllowedMarkOperator() {
        return ['levelmarker', 'bar', 'subdivider', 'whisker'];
    }

    setMarkMapping(mapping) {
        this.markMapping = (mapping || ZC_BAR_MARKS_MAPPING).call(this);
    }

    getDefaultMarkParser() {
        let operators = this.getAllowedMarkOperator(),
            markMapping = this.markMapping,
            customOperator = this.markOperator
        this._mark = executeMarkOperators.call(this, this.data, markMapping, operators, customOperator);
    }

    render(parent) {
        let holderIndex = pick(this.builderC.dynamicSeriesPosition, this.builderC.datasetIndex);
        this.placeHolder = createElement(
            'g',
            `.barseries_${holderIndex}`,
            [{ data: this.data, index: this.conf.renderingOrderIndex() }],
            parent
        );
        this._mark.map(d => {
            renderDom(d.mark, d.holder || this.placeHolder, this.chart.viewFlow);
        });
    }

    destroy() {
        this.container = this.placeHolder = this._mark = null;
    }
}
