// @zohocharts/zc/src/bar/bar-subdivider-marks.js
import { ParserRegistry } from '../core/parser-registry';

export function _subseriesDividerMark(data) {
    let _this = this;
    let currentData = data.data;

    let _subdividerMark = {
        mark: 'g',
        selector: '.subdividerseries',
        data: function() {
            return data.disabled || !_this.builderC.stacked ? [] : data.data;
        },
        properties: {
            attrs: {
                class: (d, i) => `subdividerseries subdividerseries_${i}`
            }
        },
        children: [
            {
                mark: '_rect',
                selector: '.zc-bar',
                data: function(d, i) {
                    return i > 0 ? _this.helpers.filterData(d) : [];
                },
                config: {
                    update: {
                        d: d => _this.helpers.position.subdivider.end(d)
                    }
                },
                properties: {
                    styles: {
                        fill: 'white',
                        opacity: 0.7,
                        'shape-rendering': 'crispedges'
                    }
                }
            }
        ]
    };

    return _subdividerMark;
}

ParserRegistry.marks('bar.subdivider', _subseriesDividerMark);
