// @zohocharts/zc/src/bar/bar-whisker-highlight.js
import { BarHighlighter } from './bar-highlight';
import { _super, defined, inherit } from '../utils/zcutils';
import { RECT_ELEMENT } from '../core/element-constants';
import { appendEle } from '../utils/zcsvgutils';
import { addSVGmarker, getSegmentsForSymbol } from '../utils/svg-symbols';
import { whiskerSize } from '../boxplot/whiskers';
import { patternFillEffect } from '../events/visualEffects/highlight/pattern-fill';
import { HIGHLIGHT_CONTEXT } from '../events/visualEffects/highlight/highlight-context';

export function BarWhiskerHighlighter(chart, renderer) {
    this.chart = chart;
    this.renderer = renderer;
}
inherit(BarWhiskerHighlighter, BarHighlighter);
var proto = BarWhiskerHighlighter.prototype;

proto._defaultH = function(point) {
    if (point._type == 'segment') {
        return [
            {
                shape: 'whisker',
                data: proto.data,
                segments: proto.segments,
                path: this.path,
                context: this.renderingContext,
                translate: this.translateContext
            }
        ];
    } else {
        return [
            {
                shape: RECT_ELEMENT,
                path: _super(this).path,
                data: _super(this).data,
                context: _super(this).renderingContext,
                translate: _super(this).translateContext,
                styles: patternFillEffect
            }
        ];
    }
};
proto.renderingContext = function() {
    return HIGHLIGHT_CONTEXT.get('scatter').call(this, this.chart);
};

proto.segments = function(point) {
    var chart = this.chart,
        renderer = this.renderer,
        rp = renderer.commonRendererProp().params;
    return function() {
        if (!chart.cache.data[point._key]) {
            chart.cache.data[point._key] = {};
        }
        var cache = chart.cache.data[point._key],
            transPoint;
        if (!defined(cache._segments)) {
            var _data = point._data();
            var symbolSegment = getSegmentsForSymbol(
                'horizontal-whisker',
                rp.whiskers.size,
                _data._x,
                _data._y,
                whiskerSize(point, rp.yscale, null, null, null, rp.whiskers.dataindex)
            );
            cache._segments = symbolSegment;
        }
        return point;
    };
};

proto._options = function(point) {
    var chart = this.chart,
        renderer = this.renderer,
        rp = renderer.commonRendererProp().params;

    return rp.whiskersOptionsMerged.hoveroptions;
};

proto.whiskerY = function(point) {
    var renderer = this.renderer,
        rp = renderer.commonRendererProp().params,
        whiskerIndex = rp.whiskers.dataindex;
    return (rp.yscale(point[whiskerIndex][0]) + rp.yscale(point[whiskerIndex][1])) / 2;
};
proto.path = function(point, model) {
    point = (point._data && point._data()) || point;
    var chart = this.chart,
        renderer = this.renderer,
        rp = renderer.commonRendererProp().params,
        cp = renderer.commonOption;

    var _hoveroptions = this._options(point);
    var context = model.context.call(this),
        translate = model.translate.call(this),
        { seriesIndex, paneid } = point,
        pathContext = appendEle(
            context,
            'g',
            [1],
            '.',
            'class',
            `highlight${point.paneid}${seriesIndex}${point.itemIndex}${model.type}` || '_'
        ),
        highlightEffect = 'strokeEnlarge',
        mPosition = rp.axisrotated
            ? [this.whiskerY(point), point._y + cp.barWidth / 2]
            : [point._x + cp.barWidth / 2, this.whiskerY(point)],
        wSize = whiskerSize(point, rp.yscale, null, null, null, rp.whiskers.dataindex),
        mSize;

    if (typeof _hoveroptions.size === 'string' && _hoveroptions.size.indexOf('%') > -1) {
        //if user gives size as percent, '75%'
        _hoveroptions.size = mSize = ((parseFloat(_hoveroptions.size) / 100) * cp.barWidth) / 10;
    }
    addSVGmarker(
        pathContext,
        null,
        _hoveroptions,
        {
            color: _hoveroptions.color,
            symbolSize: {
                inner: 0,
                outer: 1
            },
            symbol: 'horizontal-whisker',
            transform: mPosition,
            rotate: rp.axisrotated ? 90 : 0,
            imageSize: 15,
            size2d: wSize
        },
        this.chart
    );
    context.style('display', '');
    context.attr('transform', `translate(${translate[0]},${translate[1]})`);
    return pathContext;
};
