// @zohocharts/zc/src/bar/bar-parser-helpers.js
import { d3_values } from '../utils/zcutils';
import { ORDINAL_DATATYPE } from '../core/datatype-constants';
import { DataProcessor } from '../core/data-processor';
import { applyClippathToLivePlot } from '../plot/dynamic-live-update/dynamic-plot-clippath';
import { plot } from '../core/main-zc-init';

export let bar_parser_helpers = function() {
    let chart = this.chart,
        processor = chart.dataObject,
        dataset = chart.dataset;
    let helpers = {
        column: (type, order) => {
            let columnIdx = dataset.getColumnIdx(type, order);
            return chart.systemConf.metadata.columns[columnIdx];
        },
        datatype: (type, order) => dataset.getDataType(type, order),
        filterData: data => {
            let subseriesIdx = data.subseriesIdx || 0,
                liveUpdate = this.builderC.liveUpdate; //TODO if flatten data calculated first, we can remove this property
            if (this.seriesdataAfterUpdate[subseriesIdx] && !liveUpdate) {
                return this.seriesdataAfterUpdate[subseriesIdx];
            }

            if (liveUpdate) {
                applyClippathToLivePlot(data, this.placeHolder, this, chart);
                return DataProcessor.helpers.getDynamicData(data, this.chart);
            }

            //updateseriesdataby categories
            //renderVisiblePointsOnly
            //xAxisReverseed

            var datatypeY = chart.dataset.getDataType('y', this.commonOption.yaxiscolumnorder),
                xCategories = this.scales.x.domain(),
                yCategories = datatypeY == ORDINAL_DATATYPE ? this.scales.y.domain() : null,
                dataProcessorHelpers = DataProcessor.helpers;
            var serData = processor.updateSeriesdatabyCategories(
                d3_values(data),
                xCategories,
                yCategories,
                this.commonOption.yaxiscolumnorder,
                true,
                subseriesIdx
            );
            if (chart.renderVisiblePointsOnly) {
                serData = dataProcessorHelpers.filterPointsInViewport(chart, serData, 'bar', 'x', this.rendererIndex, subseriesIdx || 0);
            }
            this.seriesdataAfterUpdate[subseriesIdx || 0] = serData;
            let dataUpdate = this.seriesdataAfterUpdate[subseriesIdx || 0];
            this.oldRenderer.seriesdataAfterUpdate[subseriesIdx || 0] = serData; //update in existing renderer
            this.builderC.xcategoriesReversed = xCategories;
            if (this.builderC.xaxisreversed) {
                serData = serData.slice(0).reverse();
                this.builderC.xcategoriesReversed = xCategories.slice(0).reverse();
            }
            let showLevelMarker = this.builderC.levelMarker && this.builderC.levelMarker._zc_enabled;
            if (showLevelMarker && this.builderC.stacked && xCategories.length != dataUpdate.length) {
                plot.render.bulletHelpers.addDummyDataForLevelmarker(dataUpdate, xCategories, dataset, this.commonOption.yaxiscolumnorder);
            }

            return this.seriesdataAfterUpdate[subseriesIdx || 0];
        },
        position: {},
        styles: {},
        styleComponent: {}
    };
    return helpers;
};
