// @zohocharts/zc/src/bar/bar-marker-helpers.js
import { MarkerParser } from '../utils/marker-parser';
import { defined, invokeFunction, pick, validateNullVal } from '../utils/zcutils';
import { mathMin } from '../core/zc-constants';
import { isWithinColorRange } from '../utils/colorutils';
import { isMultiColoring, getStrokeColor, getFillColor } from '../utils/colorutils';

export function getBaseMark4BarMarker(type, data, options, order, prefix) {
    let _this = this,
        cType = _this.type;
    let markerParser = getMarkerParserForBar(_this, type, options, order, prefix);
    let baseG = {
        mark: 'g',
        selector: `.markers${order}`,
        data: function(d) {
            return d;
        },
        properties: {
            attrs: {
                transform: `translate(${_this.chart.plotarea.left},${_this.chart.plotarea.top})`
            }
        },
        children: [
            {
                mark: 'g',
                data: function(d) {
                    return data || _this.helpers.filterData(d);
                },
                properties: {

                    attrs: {
                        transform: function(d) {
                            let axisrotated = _this.chart.axes.rotated,
                                x = _this.helpers.position[type].end(d).x,
                                y = _this.helpers.position[type].end(d, 'y', order).y,
                                rotate = _this.helpers.position[type].end(d).rotate;
                            if(isNaN(y)){
                                return null;
                            }
                            var transform = `translate(${axisrotated ? y : x},${axisrotated ? x : y})`;
                            return cType == 'gantt' ? transform : transform + `rotate(${axisrotated ? 90 : 0})`
                            // if(cType == 'gantt') {
                            //     return `translate(${axisrotated ? y : x},${axisrotated ? x : y})`;
                            // }
                            // return `translate(${axisrotated ? y : x},${axisrotated ? x : y}) rotate(${axisrotated ? 90 : 0})`;
                        }
                    },
                    styles: {
                        display: function(d){
                            if(_this.helpers.styles[type] && _this.helpers.styles[type].display){
                                return _this.helpers.styles[type].display(d);
                            }
                            return '';  
                        }
                    }
                },
                children: [{ ...markerParser}]
            }
        ]
    };
    return baseG;
}

function getMarkerParserForBar(_this, type, options, order, prefix) {
    var renderer = _this.oldRenderer,
    cp = renderer.commonOption;

    var rp = renderer.commonRendererProp().params,
        chartObj = rp.chartObj,
        commonPlotOptions = rp.commonPlotOptions,
        color = rp.color,
        cType = cp.charttype,
        isGantt = cType == 'gantt',
        multiColoring = isMultiColoring(commonPlotOptions, chartObj);

    var clr = multiColoring ? function(d, i){
                const element = this;
                const parent = element.parentElement;
                const Sparent = parent.parentElement;
                const children = Array.from(Sparent.children);
                const index = children.indexOf(parent);
                return chartObj.seriesColor[index % chartObj.seriesColor.length]
            } : color;

    let options_cloned = { ...options };
    /*
    UPDATE STROKECOLOR & STROKEWIDTH IN OPTIONS
     */
    options_cloned.outerStrokeColor = d => {
        if (!isWithinColorRange(_this.chart.dataset.getClr(d), _this.chart)) {
            return 'transparent';
        } else {
            return options.outerStrokeColor ? options.outerStrokeColor : clr;
        }
    };
    options_cloned.strokeWidth = options.strokeWidth || 0; //DOC_BULLET_11

    /*
    HIDE INNER MARKER ON FOLLOWING CONDITIONS
     */
    let innerData;
    if (prefix == 'inner') {
        let hasImage = options.imageUrl != null && options.imageUrl !== '',
            innerColor = options.innerFillColor || options.color;
        innerData =
            hasImage || (options.innerStrokeWidth === 0 && (options.innerFillOpacity === 0 || innerColor === 'transparent')) ? [] : [1];
    }

    let _markerParser = MarkerParser({
        data: function(d) {
            return prefix == 'inner' ? innerData : [d];
        },
        prefix: prefix || 'outer',
        selection: prefix ? `.${prefix}` : '.outer',
        config: {
            x: 0,
            y: 0,
            rotate: d => _this.helpers.position[type].end(d).rotate,
            ...options_cloned,
            shape: options.symbol,
            color: isGantt ? clr : null,
            size: (d, i) => _this.helpers.position[type].end(d, 'size', order, prefix)
        },
        context: _this.chart.getPatternSvg(),
        defsPath: _this.chart.defsLocation
    });

    if (options.animation && options.animation.enabled) {
        delete _markerParser.config.strokeOpacity;
        _markerParser.config.enter = {
            strokeOpacity: 0
        };
        _markerParser.anime = {
            update: {
                duration: options.animation.duration,
                delay: options.animation.duration,
                strokeOpacity: options.outerStrokeOpacity
            }
        };
    }
    return _markerParser;
}

export function getBarMarkerSize(type, prefix) {
    //Reset old config.size
    let _config = this.builderC[type];
    if (_config._size !== undefined) {
        _config.size = _config._size;
    } else {
        _config._size = _config.size || null;
    }
    let DIM = 10,
        symbolSize = { outer: 1, inner: 0.6 },
        config = this.builderC[type],
        hasImage = config.imageUrl != null && config.imageUrl !== '',
        size = config.size;
    size = defined(size) ? size : this.commonOption.levelMarkerWidth - 2 * validateNullVal(config.padding, this.commonOption.padding / 2);
    if (defined(config.size) && typeof size === 'string' && size.indexOf('%') > -1) {
        //if user gives size as percent, '75%'
        size = ((parseFloat(size) / 100) * this.commonOption.levelMarkerWidth) / 10;
    }
    if (defined(config.size) && type == 'targetMarker') {
        size = mathMin(size, this.scales.x.bandwidth() / 10);
    }
    this.builderC[type].size = defined(config.size) ? size : size / 10;
    size = pick(config.size, 1);
    // let imgSize = (id != null ? size : size * symbolSize.outer) * DIM;
    let outersize = size * symbolSize[prefix || 'outer'] * DIM;
    return outersize;
}
