// @zohocharts/zc/src/bar/bar-marks.js
import { d3_map, d3_select } from '../core/d3-constants';
import { ParserRegistry } from '../core/parser-registry';
import { roundedRect } from '../utils/zcsvgutils';
import { defined, getBorderRadius, isUndefined, simpleClone } from '../utils/zcutils';

export function _mainBar(data) {
    let KEY = this.KEY || 'bar';
    let _this = this;
    let currentData = data.data,
        builderC = this.builderC,
        anime_config = builderC.animationOptions ?? {},
        racing = this.chart.racing && this.chart.racing.syncTransition,
        morphing = this.chart.systemConf.chart.plot.morph.enabled && this.chart.oldSeriesInfo,
        liveUpdate = builderC.liveUpdate,
        attrs = {},
        renderer = this.chart.renderer[builderC.rendererIndex],
        seriesIdx = renderer.commonRendererProp().datasetIndex();

    let _styles = this.helpers.styles[KEY] ?? {};

    let enterMarks, updateMarks, exitMarks;
    /*
    Default enter marks
     */
    enterMarks = {
        d: d => _this.helpers.position[KEY].start(d)
    };

    if (racing) {
        KEY = 'bar_racing';
        enterMarks = {
            d: d => _this.helpers.position[KEY].start(d)
        };
        updateMarks = [
            {
                transition: racing,
                d: d => _this.helpers.position[KEY].end(d)
            }
        ];

        exitMarks = {
            transition: racing,
            d: d => _this.helpers.position[KEY].exit(d)
        };
    } else if (morphing) {
        updateMarks = [
            {
                duration: anime_config._duration,
                d: {
                    property: 'attrTween',
                    to: function(d, i) {
                        let interpolator = _this.helpers.position.bar_morphing.interpolator[i];
                        return function(t) {
                            return _this.helpers.position.bar_morphing.line_d3(
                                interpolator.map(function(p) {
                                    return p(t);
                                })
                            );
                        };
                    }
                }
            },
            {
                duration: 0,
                d: d => _this.helpers.position[KEY].end(d)
            }
        ];
    } else if (liveUpdate) {
        updateMarks = [
            {
                duration: _this.phase == 'redraw' ? anime_config._duration : 0,
                d: d => _this.helpers.position[KEY].end(d)
            }
        ];
        exitMarks = { 
            duration: anime_config._exitduration,
            d: d => _this.helpers.position[KEY].exit(d)
        };
    } else {
        updateMarks = [
            {
                duration: _this.phase == 'redraw' ? _this.chart.phase == 'legendFilter' ?
                    anime_config._duration - anime_config._exitduration : anime_config._duration : 0,
                delay: _this.chart.phase == 'legendFilter' ? anime_config._exitduration : anime_config._delay,
                d: d => _this.helpers.position[KEY].stop_1(d)
            },
            {
                duration: anime_config._duration,
                delay: anime_config._delay,
                easing: anime_config.easingType,
                d: (d, i) => _this.helpers.position[KEY].end(d)
            }
        ];

        exitMarks = {
            duration: anime_config._exitduration,
            d: d => _this.helpers.position[KEY].exit(d)
        };
    }

    if (this.chart.dataObject.hierarchical) {
        attrs = {
            levelbycolumn: d => d.levelByColumn,
            levelbyrow: d => d.levelByRow
        };
    }

    let _mainPath = {
        mark: 'g',
        selector: '.subseries',
        data: function() {
            return data.disabled ? [] : data.data;
        },
        properties: {
            attrs: {
                class: (d, i) => {
                    d.subseriesIdx = i;
                    return `subseries subseries_${i}`;
                }
            }
        },
        anime: {
            exit: {
                delay: anime_config._exitduration
            }
        },
        children: [
            {
                mark: '_rect',
                selector: '.zc_bar',
                data: function(d, i) {
                    return data.disabled ? [] : _this.helpers.filterData(d);
                },
                dataAccessor: _ => _this.chart.dataset.getX(_),
                config: {
                    enter: enterMarks,
                    strokeWidth: _styles.strokeWidth,
                    dashStyle: _styles.strokeDasharray,
                    lineCap: 'none'
                },
                anime: {
                    update: updateMarks,
                    exit: exitMarks
                },
                properties: {
                    attrs,
                    styles: {
                        fill: (d, i) => _styles.fill(d, i),
                        stroke: (d, i) => _styles.stroke(d, i),
                        'fill-opacity': _styles.fillOpacity,
                        opacity: _styles.opacity
                    }
                }
            }
        ]
    };
    return _mainPath;
}

ParserRegistry.marks('bar.default', _mainBar);

// export function _mainBar(data) {
//     let _this = this;
//     let currentData = data.data;
//
//     let _mainPath = {
//         mark: 'g',
//         selector: '.subseries',
//         data: data.data,
//         properties: {
//             attrs: {
//                 class: (d, i) => `subseries subseries_${i}`
//             }
//         },
//         children: [
//             {
//                 mark: '_rect',
//                 selector: '.zc-bar',
//                 data: function (d, i) {
//                     return _this.helpers.filterData(d);
//                 },
//                 config: {
//                     enter: {
//                         y: _this.chart.plotarea.height
//                     },
//                     x: (d) => _this.helpers.position.bar.end(d, 'x'),
//
//                     // y: (d) => _this.helpers.position.bar.end(d, 'y'),
//                     width: (d) => _this.helpers.position.bar.end(d, 'width'),
//                     // height: (d) => _this.helpers.position.bar.end(d, 'height'),
//                     radius: _this.helpers.position.bar.end(null, 'radius'), //TODO If border is available as function it would be better
//                     strokeWidth: _this.helpers.styles.bar.strokeWidth
//                 },
//                 properties: {
//                     styles: {
//                         fill: (d, i) => _this.helpers.styles.bar.fill(d, i),
//                         stroke: (d, i) => _this.helpers.styles.bar.stroke(d, i),
//                         'stroke-dasharray': _this.helpers.styles.bar.strokeDasharray,
//                         'fill-opacity': _this.helpers.styles.bar.fillOpacity
//                     }
//                 },
//                 anime: {
//                     update: {
//                         duration: 500,
//                         y: { to: (d) => _this.helpers.position.bar.end(d, 'y') },
//                         height: (d) => _this.helpers.position.bar.end(d, 'height')
//                     },
//                     exit: {
//                         duration: 200,
//                         y: _this.chart.plotarea.height,
//                         height: 0
//                     }
//                 }
//             }
//         ]
//     };
//     return _mainPath;
// }

// export function _defaultBarMark(data) {
//     let _this = this;
//     let ____dummyOptions = {
//         scales: {
//             x: '',
//             y: '',
//             fillColor: ''
//         },
//         stack: true,
//         rotated: false,
//         type: 'bullet',
//         sliding: '',
//         rp: {
//             hasNegValue: '',
//             threshold: ''
//         }
//     };
//     let currentData = data.data;
//
//     let _barMark = {
//         mark: 'g',
//         selector: '.subseries',
//         data: data.data,
//         children: [
//             {
//                 mark: '_rect',
//                 selector: '.zc-bar',
//                 data: function(d, i) {
//                     return _this.helpers.filterData(d);
//                 },
//                 config: {
//                     x: d => _this.helpers.position.barEnd(d, 'x'),
//                     y: d => _this.helpers.position.barEnd(d, 'y', 0),
//                     width: d => _this.helpers.position.barEnd(d, 'width'),
//                     height: d => _this.helpers.position.barEnd(d, 'height')
//                 },
//                 properties: {
//                     styles: {
//                         fill: (d, i) => _this.helpers.styles.fill(d, i),
//                         stroke: '',
//                         'stroke-width': '',
//                         'stroke-dasharray': '',
//                         'fill-opacity': ''
//                     }
//                 }
//             }
//         ]
//     };
//     return _barMark;
//
//     return function(_data) {
//         // let currentData = data.data;
//         // let _barMark = {
//         //     mark: {
//         //         mark: 'g',
//         //         selector: '.subseries',
//         //         data: data.data
//         //     },
//         //     children: [
//         //         {
//         //             mark: 'rect-path',
//         //             selector: '.zc-bar',
//         //             data: function(d, i) {
//         //                 return this.helpers.filterData(data);
//         //             },
//         //             properties: {
//         //                 x: d => this.helpers.position.barEnd(d, 'x'),
//         //                 y: d => this.helpers.position.barEnd(d, 'y', 0),
//         //                 w: d => this.helpers.position.barEnd(d, 'width'),
//         //                 h: d => this.helpers.position.barEnd(d, 'height'),
//         //                 fill: 'blue'
//         //             }
//         //         }
//         //     ]
//         // };
//         // let _barMark = {
//         //     mark: {
//         //         mark: 'g',
//         //         // selector: '.barseries_i', //i should be dynamic
//         //         selector: '.subseries_i', //i should be dynamic
//         //         selector: function(cache) {
//         //             return cache.dataIndex;
//         //         },
//         //         data: function(d, i) {
//         //             return data.disabled ? [] : [d];
//         //         },
//         //         children: [
//         //             {
//         //                 mark: 'rect-path',
//         //                 selector: '.zc-bar',
//         //                 data: function(d, i) {
//         //                     //dataUpdate
//         //                     // let seriesdata = d.data[i];
//         //                     let seriesdata = this.helpers.filterData(data);
//         //                     //updateseriesdataby categories
//         //                     //renderVisiblePointsOnly
//         //                     //xAxisReverseed
//         //                     return options.dataset.getX(d);
//         //                 },
//         //                 properties: {
//         //                     enter: {
//         //                         //line: 595
//         //                         x: this.helpers.position('x'),
//         //                         y: this.helpers.position('y'),
//         //                         w: ____dummyOptions.rotated ? 0 : this.helpers.position('width'),
//         //                         h: ____dummyOptions.rotated ? this.helpers.position('height') : 0,
//         //                         //line: 620
//         //                         stroke: options.stroke,
//         //                         'stroke-width': options.borderSize,
//         //                         'stroke-dasharray': options.dashStyle,
//         //                         fill: options.fill,
//         //                         fill: options.fillOpacity
//         //                     },
//         //                     update: {
//         //                         y: {
//         //                             from: this.__config___.y || 'init',
//         //                             to: ''
//         //                         },
//         //                         x: {},
//         //                         y: this.helpers.position('y'),
//         //                         height: this.helpers.position('height'),
//         //                         animate: {
//         //                             //if morphing line: 753
//         //                             attrTween: {
//         //                                 d: (d, ele) => {
//         //                                     return function(t) {
//         //                                         return d3_line()(
//         //                                             ____dummyOptions.interpolator.map(function(p) {
//         //                                                 return p(t);
//         //                                             })
//         //                                         );
//         //                                     };
//         //                                 }
//         //                             },
//         //                             //normal bar line:802
//         //                             delay: i => (____dummyOptions.sliding ? 0 : i * 10),
//         //                             d: {
//         //                                 from: '',
//         //                                 to: () => this.helpers.position('all') //'all' I meant all attrs x,y,w,h
//         //                             }
//         //                         }
//         //                     },
//         //                     exit: {
//         //                         //line: 351
//         //                         d: (d, ele) => {
//         //                             return {
//         //                                 x: this.helpers.position('x'),
//         //                                 y: this.helpers.position('y'),
//         //                                 w: ____dummyOptions.rotated ? 0 : this.helpers.position('width'),
//         //                                 h: ____dummyOptions.rotated ? this.helpers.position('height') : 0
//         //                             };
//         //                         }
//         //                     }
//         //                 }
//         //             },
//         //             {
//         //                 //line:906
//         //                 mark: 'g',
//         //                 selector: '.subdividerseries_',
//         //                 data: function(d, i) {
//         //                     return data.disabled || !(i > 0 && options.stack) ? [] : [d];
//         //                 },
//         //                 children: [
//         //                     {
//         //                         name: 'UNIQUE_0', //merge properties or marks by names or any other better way
//         //                         mark: 'rect-path',
//         //                         selector: '.zc-bar',
//         //                         data: function(d) {
//         //                             return options.dataset.getX(d);
//         //                         },
//         //                         properties: {
//         //                             attrs: {
//         //                                 x: this.helpers.position('x'),
//         //                                 y: this.helpers.position('y'),
//         //                                 w: ____dummyOptions.rotated ? 1 : this.helpers.position('width'),
//         //                                 h: ____dummyOptions.rotated ? this.helpers.position('width') : 1
//         //                             },
//         //                             styles: {
//         //                                 fill: 'white',
//         //                                 opacity: 0.7,
//         //                                 'shape-rendering': 'crispedges'
//         //                             }
//         //                         }
//         //                     }
//         //                 ]
//         //             },
//         //             {
//         //                 /*
//         //                 Bar, bullet, gantt - level marker
//         //                  */
//         //                 //line:302
//         //                 mark: 'g',
//         //                 selector: options.stack ? 'levelMarkerSeries_' : '.subseries_levelMarker_',
//         //                 data: d => (options.levelMarker ? [d] : []),
//         //                 children: [
//         //                     {
//         //                         mark: 'rect-path',
//         //                         selector: '.zc-levelMarker',
//         //                         data: d => options.dataset.getX(d),
//         //                         properties: {
//         //                             enter: {
//         //                                 d: () => {
//         //                                     return {
//         //                                         //line: 512
//         //                                         x: this.helpers.levelMarkerPosition('x'),
//         //                                         y: this.helpers.levelMarkerPosition('y'),
//         //                                         w: this.helpers.levelMarkerPosition('width'),
//         //                                         h: this.helpers.levelMarkerPosition('height')
//         //                                     };
//         //                                 },
//         //                                 //line: 457
//         //                                 fill: options.fill
//         //                             },
//         //                             exit: {
//         //                                 animate: {
//         //                                     d: {
//         //                                         to: () => {
//         //                                             //line: 313
//         //                                             if (options.type == 'gantt') {
//         //                                                 return 'd'; // dummy
//         //                                             } else {
//         //                                                 return {
//         //                                                     x: this.helpers.levelMarkerPosition('x'),
//         //                                                     y: this.helpers.levelMarkerPosition('y'),
//         //                                                     w: this.helpers.levelMarkerPosition('width'),
//         //                                                     h: this.helpers.levelMarkerPosition('height')
//         //                                                 };
//         //                                             }
//         //                                         }
//         //                                     }
//         //                                 }
//         //                             }
//         //                         }
//         //                     }
//         //                 ]
//         //             },
//         //             {
//         //                 /*
//         //                 Bullet - target marker
//         //                 Gantt - connector marker
//         //                 waterfall - connector, comparison
//         //                 bar, barrange, boxplot - whisker, median
//         //                  */
//         //             }
//         //         ]
//         //     }
//         // };
//     };
// }

// User data:
// {
//     "columns": {
//     "market": {
//         "name": "Market",
//             "dataIndex": 0,
//             "dataType": "ordinal"
//     },
//     "sales":{
//         "name": "Sales",
//             "dataIndex": 1,
//             "dataType": "numeric"
//     }
// },
//     "encode":{
//     "x": "market",
//         "y": "sales"
// },
//     "series":{
//     "data": [
//         ["central", 10000],
//         ["east", 20000],
//         ["west", 30000]
//     ],
//         "type": "bar"
// }
// }
//
// Parser data:
// {
//     "columns": {
//     "market": {
//         "name": "Market",
//             "dataIndex": 0,
//             "dataType": "ordinal"
//     },
//     "sales":{
//         "name": "Sales",
//             "dataIndex": 1,
//             "dataType": "numeric"
//     }
// },
//     "series": [{
//     "data":[
//         ["central", 10000],
//         ["east", 20000],
//         ["west", 30000]
//     ],
//     "mark": [
//         {
//             "mark": "rect-path",
//             "encode":{
//                 "x": "market",
//                 "y": "sales",
//                 "color": "",
//             p    "size": ""
//             },
//worst case
//"transition": {
//
// },
//             "animate":{
//custom hooks adding any d3 transition or selection API
//                 "enter": {
//                     //on start, end // what is the use here? is there any other way we can use this
//                     "transition":{
//                         "delay": ""
//                     },
//                     "attrs":{
//                         "y": {
//                             "from": "", //from element may be object or function or self element or we can ignore this pro
//                             "to": ""
//                         }
//                     }
//
//                 },
//                 "exit": {
//                     "height": {
//                         "from": "",
//                         "to": ""
//                     }
//                 }
//             },
//             "enter": {
//                 "x": "",
//                 "y": "",
//                 "w": "",
//                 "h": ""
//             }
//         }
//     ]
// }]
// }
