// @zohocharts/zc/src/bar/barchart.js
//$Id$
//barchart.js
import { LabelRegistry } from '../label/label-registry';
import { svg_plot_renderer, plot, chartcategory, axis, stack } from '../core/main-zc-init';
import { d3_select, d3_line } from '../core/d3-constants';
import { utils, mergeJSON, toArray, isTrue, d3_values, defined, isArray, pick } from '../utils/zcutils';
import { ORDINAL_DATATYPE } from '../core/datatype-constants';
import { gradient_helpers } from '../effects/gradient-helpers';
import { appendEle, parseSVGPath, roundedRect } from '../utils/zcsvgutils';
import { isMultiColoring, getStrokeColor, getFillColor } from '../utils/colorutils';
import { getAnimationConf } from '../utils/animationutils';
import { getBarCommonOption } from './barRendererConf';
import { getDashArray } from '../utils/dashstyles';
import { effects } from '../effects/zceffects';
import { VERTICALALL_ANIMATION } from '../core/animation-constants';
import { mathMin } from '../core/zc-constants';
// import { getStackGroupIndex } from '../core/data-processor/transforms/stack/stack-grouping';
import { getMarkerConfig } from '../boxplot/markerOptions';
import { BarHighlighter } from './bar-highlight';
import { BulletHighlighter } from '../bullet/bullet-highlight';
import { BoxplotHighlighter } from '../boxplot/boxplot-highlight';
import { GanttHighlighter } from '../gantt/gantt-highlight';
import { DataProcessor } from '../core/data-processor';
import { getChartTypeByIndex } from '../core/charttype-helpers';
import { setOpacityForTickFilterElements } from '../plot/svg/set-pivot-attr';
import { getSegmentsForSymbol } from '../utils/svg-symbols';
import { whiskerSize } from '../boxplot/whiskers';
import { BarWhiskerHighlighter } from './bar-whisker-highlight';
import { zcMorph } from '../morphing/zc-morping';
import { getRawColor } from '../utils/colorutils';
import { getOffsetValue } from '../plot/morph/morphTools';
import { BarRenderer } from './bar-parser';
import { BoxplotRenderer } from '../boxplot/boxplot-parser';
import { BulletRenderer } from '../bullet/bullet-parser';
import { getAxisData } from '../axis/axisutils'; 
import { isFacet } from '../axis/hierarchy-axis-utils';
import { ParserRegistry } from '../core/parser-registry';
import ArcHighlighter from '../highlighter/arc-highlighter';
import { GanttRenderer } from '../gantt/gantt-parser';
// import { RadialBarParser } from './parser/radial/radialbar-parser';

/**
 * @returns {chart} bar renderer.
 */

svg_plot_renderer.bar = function() {
    var commonRendererProp,
        bandwidth1,
        bandwidth2,
        yscale,
        barDatasetIndex,
        isredraw = false,
        shiftInfo,
        transitionDuration,
        dataProcessorHelpers = DataProcessor.helpers,
        newBarParser = false,
        axisType,
        parser;

    function chart(selection) {
        selection.each(function(data) {
            var container = d3_select(this);
            chart.update = function() {
                container.transition().call(chart);
            };

            //renderer config values
            var rp = commonRendererProp.params,
                chartObj = rp.chartObj,
                dataset = chartObj.dataset,
                xscale = rp.xscale,
                scaleType = (axisType = dataset.getColumn('x')?.scaleType),
                stacked = rp.stacked,
                index = rp.datasetIndex,
                renderingOrder = rp.renderingOrderIndex,
                legendOrder = rp.legendOrderIndex,
                rendererIndex = rp.rendererIndex,
                axisrotated = rp.axisrotated,
                commonPlotOptions = rp.commonPlotOptions,
                color = rp.color,
                paneColor = rp.paneColor,
                labelQueue = rp.labelQueue,
                hasNegValue = rp.hasNegValue,
                yaxiscolumnorder = rp.yaxiscolumnorder,
                dataset = rp.dataset,
                systemconf = chartObj.systemConf,
                plotarea = chartObj.plotarea,
                dataObject = chartObj.dataObject,
                threshold = rp.threshold,
                colorOverlay = rp.colorOverlay,
                xaxisData = getAxisData(chartObj, 'x', 0, false),
                yaxisData = getAxisData(chartObj, 'y', rp.yaxiscolumnorder, false),
                yaxisReversed = (rp.yaxisreversed = yaxisData.reversed),
                xAxisReverseed = (rp.xaxisreversed = xaxisData.reversed),
                allowDuplicateCategories = systemconf.chart.plot.allowDuplicateCategories,
                parser_data,
                data1;
            yscale = rp.yscale;

            if (chartObj.viewFlow && chartObj.viewFlow.marks._default.chart.length) {
                // let marks_plot = chartObj.viewFlow.marks._default;
                // renderdata = marks_plot;
                parser_data = data[index];
                data1 = parser_data.data;
            }
            if(chartObj.customVisual){
                chart.chartRendered = true;
                return;
            }
            var currentdata = data1 || data[index],
                data_bind = { data: currentdata, index: renderingOrder },
                defsLocation = chartObj.defsLocation,
                plotoptions = currentdata.plotoptions,
                currentseriesdata = currentdata.data,
                paneid = isFacet(chartObj) ? `${currentdata.row}_${currentdata.column}` : '0_0',
                cType_Obj = getChartTypeByIndex(chartObj, paneid, index),
                cType = cType_Obj.name,
                plotConfig = chartObj.dataSplicer.getPlotConfig(parser_data || currentdata, index),
                gvp = (rp.gvp = utils.getValueOfPlotOptions(plotConfig, commonPlotOptions)),
                gradientDim = gradient_helpers.getDimension(chartObj, index, paneid),
                patternPallete = chartObj.patternPallete && chartObj.patternPallete.length,
                noofbarDatas = chartObj.dataObject.noofBarEnabledSeries,
                isBullet = cType == 'bullet',
                isGantt = cType == 'gantt',
                isWaterfall = cType == 'waterfall',
                isBoxplot = cType == 'boxplot',
                isRangePlot = chartcategory.rangePlot.has(cType_Obj.value),
                isWhiskerPlot = chartcategory.whiskerPlot.has(cType_Obj.value),
                datatypeY = chartObj.dataset.getDataType('y', yaxiscolumnorder),
                xCategories = xscale.domain(),
                yCategories = datatypeY == ORDINAL_DATATYPE ? yscale.domain() : null,
                categories = defined(dataObject.hierarchical) ? dataObject.getCategories('x') : xCategories,
                categoryReversed = xCategories,
                dynamicPosition = rp.dynamicSeriesPosition;
            if (dynamicPosition === undefined) {
                dynamicPosition = index;
            }

            /*
                Trying bar mark parser
             */
            if (cType != 'waterfall') {
                newBarParser = true;

                if (scaleType == 'polar') {
                    let dependencies = {
                        container, 
                        chart: chartObj,
                        data: data1 || data[index],
                        renderer: chart,
                        operator: parser_data?.operator,
                        enhancer: parser_data?.enhancer,
                        config: {
                            common: commonPlotOptions,
                            series: plotConfig,
                            merged: gvp
                        },
                        rendererIndex, legendOrder, barDatasetIndex,
                        builderC: rp,
                        phase: isredraw ? 'redraw' : 'draw'
                    }
                    parser = new (ParserRegistry.parser('RadialBar'))(dependencies);
                    parser.render();

                } else {
                    rp.liveUpdate = shiftInfo;
                    barDatasetIndex = pick(parser_data && parser_data.datasetIndex, barDatasetIndex);
                    let args = {
                        type: cType,
                        scaleType,
                        chart: chartObj,
                        oldRenderer: chart,
                        data: data1 || data[index],
                        operator: parser_data ? parser_data.operator : null,
                        paneid,
                        legendOrder,
                        rendererIndex,
                        barDatasetIndex,
                        config: {
                            common: commonPlotOptions,
                            series: plotConfig,
                            merged: gvp
                        },
                        builderC: rp,
                        phase: isredraw ? 'redraw' : 'draw'
                    };
                    commonRendererProp.parser = parser = cType == 'boxplot' ? new BoxplotRenderer(args)
                        : cType == 'bullet' ? new BulletRenderer(args) 
                        : cType == 'gantt' ? new GanttRenderer(args) : new BarRenderer(args);
                    parser.initialize(container);
                    if ((!bandwidth1 || !bandwidth2) && scaleType != 'polar') {
                        //Test_515.json (Combinational chart(bar + line) where all the bar series are disabled)
                        isredraw = true;
                        chart.chartRendered = true;
                        return;
                    }
                    parser.getDefaultMarkParser();
                    parser.render(container);
                }

                /*
                HOOK for DATALABEL
                
                
                 */
                let builderC = parser.builderC,
                    seriesDataUpdate = parser.seriesdataAfterUpdate;
                if (isBoxplot) {
                    var yidx = chartObj.dataset.getColumnIdx('y', yaxiscolumnorder),
                        medianidx = builderC.median.columnindex,
                        whiskeridx = builderC.whiskers.columnindex;
                    var labelrenderer = chartObj.datalabels.renderer,
                        labelConfig = labelrenderer ? labelrenderer[index] : {};
                    var labeltype_boxplot = [];
                    if (labelConfig && labelConfig.show) {
                        var labeltype = labelConfig.labeltype;
                        var labelidx;
                        if (typeof labeltype == 'object') {
                            var indices = labeltype.get('val');
                            if (indices && indices.length == 1) {
                                //CURRENTLY MULTI CUSTOMIZED INDICES NOT SUPPORTED
                                //FOR THAT CHANGES TO BE DONE IN getCustomizedVal for parsing only particular {{val(3)}} from separator
                                labelidx = indices[0];
                                if (labelidx == medianidx) {
                                    labeltype_boxplot.push(labelidx);
                                } else if (labelidx == whiskeridx || labelidx == yidx) {
                                    labeltype_boxplot.push(labelidx, labelidx);
                                } else {
                                    labeltype_boxplot.push('others');
                                }
                            } else {
                                labeltype_boxplot.push('others');
                            }
                        } else if (labeltype == 'y') {
                            labeltype_boxplot.push(yidx, yidx);
                        } else {
                            labeltype_boxplot.push('others');
                        }

                        labeltype_boxplot.map(function(d, i) {
                            var rangeidx = d == medianidx || d == 'others' ? null : i % 2,
                                columnidx = d == 'others' ? null : d;
                            labelQueue.push({
                                renderer: chart,
                                charttype: 'bar',
                                chartObj: chartObj,
                                rendererIndex: rendererIndex,
                                paneid: paneid,
                                args: [currentdata, index, 0, rangeidx, columnidx]
                            });
                        });
                    }
                } else if (cType == 'barrange') {
                    labelQueue.push({
                        renderer: chart,
                        charttype: 'bar',
                        chartObj: chartObj,
                        paneid: paneid,
                        args: [currentdata, index, 0, 1]
                    });
                    labelQueue.push({
                        renderer: chart,
                        charttype: 'bar',
                        chartObj: chartObj,
                        paneid: paneid,
                        args: [currentdata, index, 0, 0]
                    }); //last argument in args for yrangePos
                } else {
                    if (cType == 'bullet') {
                        LabelRegistry.get('bullet').call(chartObj, builderC.targetMarker, chart, seriesDataUpdate[0], {
                            paneid,
                            rendererIndex,
                            subIndex: 0
                        }); //TODO subIndex hardcoded
                    }
                    labelQueue.push({
                        renderer: chart,
                        charttype: 'bar',
                        chartObj: chartObj,
                        rendererIndex: rendererIndex,
                        paneid: paneid,
                        args: [currentdata, legendOrder, null, null, null, rendererIndex]
                    });
                }
                return;
            }
            /*
                The End
             */
            var series = appendEle(container, 'g', [data_bind], '.', 'class', `barseries_${dynamicPosition}`),
                x = 0;

            if (!bandwidth1 || !bandwidth2) {
                //Test_515.json (Combinational chart(bar + line) where all the bar series are disabled)
                isredraw = true;
                chart.chartRendered = true;
                return;
            }

            var levelMarker = (rp.levelMarker =
                    plotoptions && plotoptions.levelMarker
                        ? mergeJSON(commonPlotOptions.levelMarker, plotoptions.levelMarker, true)
                        : commonPlotOptions.levelMarker),
                borderOptions =
                    plotoptions && plotoptions.border
                        ? mergeJSON(commonPlotOptions.border, plotoptions.border, true)
                        : commonPlotOptions.border,
                multiColoring = isMultiColoring(commonPlotOptions, chartObj);
            //Animation variables
            var animationOptions = getAnimationConf(commonPlotOptions, plotoptions, chartObj, 'bar'),
                isAnimationEnable = animationOptions.enabled,
                easingType = animationOptions.easingType,
                animationType = animationOptions.type,
                exitDuration = isAnimationEnable ? 100 : 0,
                duration;
            transitionDuration = isredraw ? (isAnimationEnable ? 300 : 0) : animationOptions.duration;
            let transition_ = chartObj.syncTransition || selection.transition().duration(transitionDuration);

            /**
             *
             * @type {{stacked, categories, axisrotated, bandwidth1, bandwidth2, barDatasetIndex, padding, yscale, hasNegValue, yaxiscolumnorder, addRangeBand, barWidth, levelMarkerWidth, levelMarkerPadding}|*}
             */
            var commonOption = getBarCommonOption(chartObj.rendererConf[rendererIndex], rendererIndex, legendOrder, paneid),
                barWidth = commonOption.barWidth,
                addRangeBand = commonOption.addRangeBand,
                getStackGroupIndex = stack.getStackGroupIndex,
                actualStackGroupIndex = getStackGroupIndex && getStackGroupIndex(dataObject.getStackGroup(), index);

            chart.getParamValue = function() {
                return chartObj.dataObject.getTransRectValue.apply(chartObj.dataObject, toArray(arguments));
            };

            const getMultiColorIndex = (d, j) => {
                if (allowDuplicateCategories) {
                    d.categoryIndex = j = categoryReversed.indexOf(dataset.getX(d));
                }
                return j;
            };

            const getGanttColor = (d, j) => {
                if (multiColoring) {
                    j = getMultiColorIndex(d, j);
                    return chartObj.seriesColor[j % chartObj.seriesColor.length];
                } else {
                    return color;
                }
            };

            var targetMarker, targetMarkerSeries, levelMarkerColor, whiskers, whiskerSeries;
            if (levelMarker) {
                //For waterfall there will be no levelmarker
                levelMarkerColor = levelMarker.color; //Saving the levelMarker colors, helps in multi categories bullet if each categories dont have same no of levels
                levelMarker.overlayRange = 'below';
            }

            if (isBullet || (isRangePlot && cType !== 'barrange')) {
                var markerPropertyName = isGantt ? 'marker' : isBoxplot ? 'median' : 'targetMarker',
                    groupName = isBoxplot ? 'threshold' : 'qualitativeGroup';

                commonRendererProp.params.targetMarker = targetMarker =
                    plotoptions && plotoptions[markerPropertyName]
                        ? mergeJSON(commonPlotOptions[markerPropertyName], plotoptions[markerPropertyName], true)
                        : commonPlotOptions[markerPropertyName];
                targetMarker = isBoxplot ? getMarkerConfig(targetMarker, animationOptions) : targetMarker;
                targetMarkerSeries = appendEle(
                    chartObj.staticPaneSVG.selectAll(`g.${groupName}`),
                    'g',
                    targetMarker.enabled === true && !currentdata.disabled ? [index] : [],
                    '.',
                    'class',
                    isGantt || isBoxplot ? `markerSeries${index}` : 'markerSeries'
                );
            }
            if (isWhiskerPlot) {
                var groupName = isBoxplot ? 'threshold' : 'qualitativeGroup';
                commonRendererProp.params.whiskers = whiskers =
                    plotoptions && plotoptions.whiskers
                        ? mergeJSON(commonPlotOptions.whiskers, plotoptions.whiskers, true)
                        : commonPlotOptions.whiskers;

                var whiskersSymbol = commonRendererProp.params.whiskers.showOn == 'edges' ? 'whiskers' : 'horizontal-whisker';
                whiskers = getMarkerConfig(whiskers, animationOptions, whiskersSymbol);

                commonRendererProp.params.whiskersOptionsMerged = whiskers;
                whiskerSeries = appendEle(
                    chartObj.staticPaneSVG.selectAll(`g.${groupName}`),
                    'g',
                    whiskers.enabled === true && !currentdata.disabled ? [index] : [],
                    '.',
                    'class',
                    `whiskerSeries${index}`
                );
            }

            var startRadius = {
                    tlx: 0,
                    tly: 0,
                    trx: 0,
                    trY: 0,
                    brx: 0,
                    bry: 0,
                    blx: 0,
                    bly: 0
                },
                borderRadius = startRadius,
                borderSize = 0,
                borderColor = null,
                borderStyle = null,
                levelMarkerRadius = startRadius;
            if (defined(borderOptions) && isTrue(borderOptions.show)) {
                borderSize = borderOptions.size;
                borderColor = borderOptions.color;
                borderStyle = borderOptions.style;
                borderRadius = chart.getParamValue('radius', null, null, null, commonOption, borderOptions.radius);
            }

            var dashStyle = getDashArray(borderStyle, borderSize == null ? 2 : borderSize);
            var fillColor = (chart.fillColorFunc = getFillColor(chartObj, index, paneid, gradientDim, null, null, null, null)),
                strokeColor = getStrokeColor(chartObj, index, paneid, rp);

            //Save seriesdata reference without null points , that can also be used in zcevents
            chart.seriesdataAfterUpdate = [];
            currentseriesdata.forEach(function(d, i) {
                chart.getBarLevelMarkerProp = isGantt ? plot.render.bulletHelpers.getLevelMarkerParams : null;
                var valuey;
                var startx, starty, startwid, starthei, endx, endy, endwid, endhei;
                var ishideable = currentdata.disabled;

                var serData = chartObj.dataObject.updateSeriesdatabyCategories(
                    d3_values(d),
                    xCategories,
                    yCategories,
                    yaxiscolumnorder,
                    true
                );
                if (chartObj.renderVisiblePointsOnly) {
                    serData = dataProcessorHelpers.filterPointsInViewport(chartObj, serData, 'bar', 'x', index, i);
                }
                chart.seriesdataAfterUpdate[i] = serData;

                if (xAxisReverseed) {
                    serData = serData.slice(0).reverse();
                    categoryReversed = categoryReversed.slice(0).reverse();
                }

                var ShiftedData;
                if (shiftInfo != null) {
                    ShiftedData = d3_values(d);
                    var co = shiftInfo.count;
                    for (var ii = 0; ii <= co; ii++) {
                        ShiftedData.splice(ii, 0, [null, null]);
                    }

                    if (isredraw) {
                        var clipX = xscale(dataset.getX(d[0]));
                        var properties = {
                                name: 'rect',
                                id: chartObj.id,
                                x: clipX,
                                y: 0,
                                width: chartObj.plotarea.width,
                                height: chartObj.plotarea.height
                            },
                            args = {
                                chartObj: chartObj
                            };
                        var clipID = effects.get(effects.svg_clips, 'rect', properties, args);
                        series.attr('clip-path', `url(${defsLocation}#${clipID})`);
                    }
                }
                var dataUpdate = shiftInfo != null ? ShiftedData : chart.seriesdataAfterUpdate[i];
                duration =
                    animationType == VERTICALALL_ANIMATION || isredraw
                        ? transitionDuration
                        : transitionDuration / (dataUpdate.length * noofbarDatas);
                var showLevelMarker =
                    levelMarker &&
                    levelMarker.enabled === true &&
                    ((stacked && noofbarDatas > 0) || !ishideable) &&
                    !chartObj.dataObject.isStackedPercentChart &&
                    (!stacked || !!chartObj.dataObject.getEncode('stack', actualStackGroupIndex)); // If specified stackgroup is disabled, remove it

                //subseries levelmarker
                var subseries_levelMarkerNode = (stacked ? container : series)
                    .selectAll(stacked ? `g.levelMarkerSeries_${actualStackGroupIndex}` : `g.subseries_levelMarker_${i}`)
                    .data(function(d) {
                        return showLevelMarker ? [d] : [];
                    });

                //Removing unnecessary 'g' on redraw ((stacked bar to bar) or (bar to stacked bar))
                (!stacked ? container : series)
                    .selectAll(!stacked ? `g.levelMarkerSeries_${actualStackGroupIndex}` : `g.subseries_levelMarker_${i}`)
                    .remove();

                subseries_levelMarkerNode
                    .exit()
                    .selectAll('path.levelMarker')
                    .transition()
                    .duration(exitDuration)
                    .attr('d', function(d, j) {
                        if (isGantt) {
                            var path = d3_select(this).attr('d'); // If the category is no more available at the categoryStore, We get the x,y from the 'd' attr by parsing
                            var points = parseSVGPath(path);
                            var x = points[0].moves[0],
                                y = points[3].moves[1];

                            var prop = plot.render.bulletHelpers.getLevelMarkerParams(d, i, j, chart, commonOption);
                            var exitx = prop.x || x,
                                exity = prop.y + (axisrotated ? 0 : prop.height * (yaxisReversed ? -1 : 1)) || y,
                                exitwid = axisrotated ? 0 : prop.width,
                                exithei = axisrotated ? prop.height : 0;
                            return roundedRect(exitx, exity, exitwid, exithei, borderRadius, yaxisReversed, axisrotated);
                        }
                        var exitx = chart.getParamValue('x', d, i, j, commonOption) || parseSVGPath(d3_select(this).attr('d'))[0].moves[0],
                            exity = yscale(yscale.domain()[0]),
                            exitwid = axisrotated ? 0 : chart.getParamValue('width', d, i, j, commonOption),
                            exithei = axisrotated ? chart.getParamValue('height', d, i, j, commonOption) : 0;
                        return roundedRect(exitx, exity, exitwid, exithei, borderRadius, yaxisReversed, axisrotated);
                    })
                    .remove();

                //subseries
                var subseriesNode = series.selectAll(`g.subseries_${i}`).data(function(d) {
                    return ishideable ? [] : [d];
                });

                subseriesNode
                    .exit()
                    .selectAll('path.zc_bar')
                    .transition()
                    .duration(exitDuration)
                    .attr('d', function(d, j) {
                        var path = d3_select(this).attr('d'); // If the category is no more available at the categoryStore, We get the x,y from the 'd' attr by parsing
                        var points = parseSVGPath(path);
                        var x = points[0].moves[0],
                            y = points[3].moves[1],
                            exitx = chart.getParamValue('x', d, i, j, commonOption) || x,
                            exity =
                                (axisrotated
                                    ? chart.getParamValue('y', d, i, j, commonOption)
                                    : isRangePlot
                                    ? chart.getParamValue('y', d, i, j, commonOption, null, 0)
                                    : (stacked
                                          ? yscale(d.y0)
                                          : hasNegValue != null
                                          ? yscale(hasNegValue)
                                          : yscale(0) != null
                                          ? mathMin(yscale(0), yscale.range()[0])
                                          : yscale.range()[0]) +
                                      addRangeBand() * 2) || y,
                            exitwid = axisrotated ? 0 : chart.getParamValue('width', d, i, j, commonOption),
                            exithei = axisrotated ? chart.getParamValue('height', d, i, j, commonOption) : 0;
                        return roundedRect(exitx, exity, exitwid, exithei, borderRadius, yaxisReversed, axisrotated);
                    })
                    .remove();

                var subseries = subseriesNode
                    .enter()
                    .append('g')
                    .merge(subseriesNode)
                    .attr('class', `subseries_${i}`);

                if (isWaterfall) {
                    //Connector lines
                    var subseries_connectorLines = appendEle(
                        stacked ? container : series,
                        'g',
                        stacked || !ishideable ? [1] : [],
                        '.',
                        'class',
                        'connectorLines',
                        null
                    );
                    svg_plot_renderer.waterfallHelpers.drawConnectorLines(subseries_connectorLines, dataUpdate, {
                        chart: chartObj,
                        yaxiscolumnorder: yaxiscolumnorder,
                        axisrotated: axisrotated,
                        stacked: stacked,
                        xscale: xscale,
                        yscale: yscale,
                        renderer: chart,
                        seriesindex: i,
                        plotoptions: plotoptions,
                        commonPlotOptions: commonPlotOptions
                    });

                    //Comparison lines
                    var subseries_comparisonLines = appendEle(
                        stacked ? container : series,
                        'g',
                        stacked || !ishideable ? [1] : [],
                        '.',
                        'class',
                        'comparisonLines',
                        null
                    );
                    svg_plot_renderer.waterfallHelpers.drawComparisonLine(subseries_comparisonLines, dataUpdate, {
                        chart: chartObj,
                        yaxiscolumnorder: yaxiscolumnorder,
                        axisrotated: axisrotated,
                        stacked: stacked,
                        xscale: xscale,
                        yscale: yscale,
                        renderer: chart,
                        seriesindex: i,
                        plotoptions: plotoptions,
                        commonPlotOptions: commonPlotOptions
                    });
                }

                var subseries_levelMarkerZindex = isGantt
                    ? subseries_levelMarkerNode
                          .enter()
                          .append('g')
                          .merge(subseries_levelMarkerNode)
                    : subseries_levelMarkerNode
                          .enter()
                          .insert('g', ':first-child')
                          .merge(subseries_levelMarkerNode);

                subseries_levelMarkerZindex.attr(
                    'class',
                    stacked ? `levelMarkerSeries_${actualStackGroupIndex}` : `subseries_levelMarker_${i}`
                );
                if (showLevelMarker && stacked && xCategories.length != dataUpdate.length) {
                    plot.render.bulletHelpers.addDummyDataForLevelmarker(dataUpdate, xCategories, dataset, yaxiscolumnorder);
                }

                var rect_levelMarkerNode = subseries_levelMarkerZindex.selectAll('path.levelMarker').data(dataUpdate, function(d) {
                    return isGantt ? d : dataset.getX(d);
                });

                var rect_levelMarker = rect_levelMarkerNode
                    .enter()
                    .append('path')
                    .merge(rect_levelMarkerNode)
                    .attr('class', 'levelMarker');
                rect_levelMarker
                    .styles({
                        fill: function(d, j) {
                            if (isGantt) {
                                return levelMarker.color || getGanttColor(d, j);
                                // return levelMarker.color || (multiColoring ? chartObj.seriesColor[j % chartObj.seriesColor.length] : color);
                            }
                            if (
                                !plot.render.bulletHelpers.ZC_BAR_getLevelMarkerRect ||
                                !plot.render.bulletHelpers.getBulletLevelMarkerGradient
                            ) {
                                throw 'Level marker or bullet module was not loaded';
                            }
                            // var baseClr = multiColoring ? chartObj.seriesColor[j % chartObj.seriesColor.length] : color;
                            var baseClr = getGanttColor(d, j);
                            var clr = isArray(levelMarker.color) ? levelMarker.color[0] : levelMarker.color;
                            if (patternPallete) {
                                var prop = isGantt
                                    ? plot.render.bulletHelpers.getLevelMarkerParams(d, i, j, chart, commonOption)
                                    : plot.render.bulletHelpers.ZC_BAR_getLevelMarkerRect(
                                          d,
                                          levelMarker,
                                          chartObj,
                                          commonOption,
                                          rp,
                                          i,
                                          index
                                      );
                                var x = prop.x,
                                    y = prop.y,
                                    wid = prop.width,
                                    hei = prop.height;
                                // #ZC2804 $j param
                                clr = fillColor(multiColoring ? getMultiColorIndex(d, j) : legendOrder, null, null, clr, null, null, {
                                    x: x,
                                    y: y,
                                    width: wid,
                                    height: hei
                                });
                            }

                            return (
                                plot.render.bulletHelpers.getBulletLevelMarkerGradient(
                                    d,
                                    levelMarker,
                                    chartObj,
                                    baseClr,
                                    levelMarkerColor
                                ) ||
                                clr ||
                                baseClr
                            );
                        },
                        'fill-opacity': levelMarker ? levelMarker.fillOpacity : null
                    })
                    .attr('d', function(d, j) {
                        var prop = isGantt
                            ? plot.render.bulletHelpers.getLevelMarkerParams(d, i, j, chart, commonOption)
                            : plot.render.bulletHelpers.ZC_BAR_getLevelMarkerRect(d, levelMarker, chartObj, commonOption, rp, i, index);
                        var x = prop.x,
                            y = prop.y,
                            wid = prop.width,
                            hei = prop.height;
                        if (isGantt) {
                            return roundedRect(x, y, wid, hei, levelMarkerRadius, yaxisReversed, axisrotated);
                        } else {
                            return axisrotated
                                ? roundedRect(y, x, wid, hei, borderRadius, yaxisReversed, axisrotated)
                                : roundedRect(x, y, wid, hei, borderRadius, yaxisReversed, axisrotated);
                        }
                    });
                setOpacityForTickFilterElements(rect_levelMarker, chartObj, cType);

                if (dataObject.hierarchical) {
                    rect_levelMarker.attr('levelbycolumn', function(d, i) {
                        return d.levelByColumn;
                    });
                    if (dataObject.pivot) {
                        rect_levelMarker.attr('levelbyrow', function(d, i) {
                            return d.levelByRow;
                        });
                    }
                }

                if (levelMarker) {
                    levelMarker.color = levelMarkerColor; //Saving back the levelMarker colors, helps in multi categories bullet if each categories dont have same no of levels
                }

                rect_levelMarkerNode.exit().remove();

                if (isBullet) {
                    svg_plot_renderer.ganttHelpers.addMarker(targetMarkerSeries, dataUpdate, chart, i, targetMarker, multiColoring, '');
                    LabelRegistry.get('bullet').call(chartObj, targetMarker, chart, dataUpdate, { paneid, rendererIndex, subIndex: i });
                } else if (isGantt) {
                    svg_plot_renderer.ganttHelpers.addMarker(
                        targetMarkerSeries,
                        dataUpdate,
                        chart,
                        i,
                        targetMarker.start,
                        multiColoring,
                        0
                    );
                    svg_plot_renderer.ganttHelpers.addMarker(targetMarkerSeries, dataUpdate, chart, i, targetMarker.end, multiColoring, 1);
                } else if (isWhiskerPlot) {
                    //For Median line
                    if (isBoxplot) {
                        svg_plot_renderer.ganttHelpers.addMarker(
                            targetMarkerSeries,
                            dataUpdate,
                            chart,
                            index,
                            targetMarker,
                            multiColoring,
                            ''
                        );
                    }
                    //For whiskers (minimum and maximum)
                    if (commonRendererProp.params.whiskers.showOn == 'edges') {
                        svg_plot_renderer.ganttHelpers.addMarker(whiskerSeries, dataUpdate, chart, index, whiskers, multiColoring, 0);
                        svg_plot_renderer.ganttHelpers.addMarker(whiskerSeries, dataUpdate, chart, index, whiskers, multiColoring, 1);
                    } else {
                        svg_plot_renderer.ganttHelpers.addMarker(whiskerSeries, dataUpdate, chart, index, whiskers, multiColoring, 0);
                    }
                }

                //Main rect
                var rect = subseries
                    .selectAll('path.zc_bar')
                    .attr('class', 'zc_bar')
                    .data(dataUpdate, function(d) {
                        return isGantt ? d : dataset.getX(d); // For gantt chart we should support different timeline data for the same category
                    });
                var rectEnter = rect
                    .enter()
                    .append('path')
                    .style('fill-opacity', gvp('fillOpacity'))
                    .attrs({
                        class: 'zc_bar',
                        height: function(d, j) {
                            if (chartObj.syncTransition) {
                                var param = axisrotated ? 'width' : 'height';
                                return chart.getParamValue(param, d, i, j, commonOption);
                            } else {
                                return 0;
                            }
                        },
                        d: function(d, j) {
                            var params = getInitialParams(d);
                            if (chartObj.syncTransition) {
                                if (axisrotated) {
                                    params.x = chartObj.plotarea.height + 20;
                                    params.y = chart.getParamValue('x', d, i, j, commonOption);
                                    params.width = chart.getParamValue('width', d, i, j, commonOption);
                                    params.height = chart.getParamValue('height', d, i, j, commonOption);
                                } else {
                                    if (chartObj.axes.x.reversed) {
                                        params.x = chartObj.plotarea.height + 20;
                                        params.y = chart.getParamValue('x', d, i, j, commonOption);
                                        params.width = chart.getParamValue('width', d, i, j, commonOption);
                                        params.height = chart.getParamValue('height', d, i, j, commonOption);
                                    }
                                }
                            }
                            d3_select(this).attr('y', params.y);
                            return axisrotated
                                ? roundedRect(params.y, params.x, params.width, params.height, startRadius, yaxisReversed, axisrotated)
                                : roundedRect(params.x, params.y, params.width, params.height, startRadius, yaxisReversed, axisrotated);
                        }
                    })
                    .merge(rect);

                var rectUpdate = rectEnter
                    .attrs({
                        stroke: function(d, j) {
                            // return getMultiColorIndex(d, pick(d.sortedItemIndex, j));
                            return strokeColor(multiColoring ? getMultiColorIndex(d, j) : legendOrder, d, borderColor || paneColor);
                        },
                        'stroke-width': borderSize,
                        'stroke-dasharray': dashStyle
                    })
                    .styles({
                        fill: function(d, j) {
                            if (patternPallete) {
                                return fillColor(multiColoring ? getMultiColorIndex(d, j) : legendOrder, d, null, null, null, null, {
                                    x: chart.getParamValue('x', d, i, j, commonOption),
                                    y: chart.getParamValue('y', d, i, j, commonOption),
                                    width: chart.getParamValue('width', d, i, j, commonOption),
                                    height: chart.getParamValue('height', d, i, j, commonOption)
                                });
                            }
                            // #ZC2087
                            // return getMulticolorIndex(d, pick(d.sortedItemIndex, j));
                            return fillColor(
                                multiColoring ? getMultiColorIndex(d, pick(d.sortedItemIndex, j)) : legendOrder,
                                d,
                                null,
                                paneColor
                            );
                        },
                        'fill-opacity': gvp('fillOpacity'),
                        transition: 'opacity ease 0.2s'
                    });

                setOpacityForTickFilterElements(rectUpdate, chartObj, cType);

                if (dataObject.hierarchical) {
                    rectEnter.attr('levelbycolumn', function(d, i) {
                        return d.levelByColumn;
                    });
                    if (dataObject.pivot) {
                        rectEnter.attr('levelbyrow', function(d, i) {
                            return d.levelByRow;
                        });
                    }
                }
                if (shiftInfo == null) {
                    if (chartObj.syncTransition) {
                        rectEnter
                            .transition(chartObj.syncTransition)
                            .attr('d', function(d, k) {
                                var params = getFinalParams(d, k);
                                if (isNaN(params.x) || isNaN(params.y)) {
                                    params.x = -1;
                                    return null;
                                }
                                return roundedRect(
                                    params.x,
                                    params.y,
                                    params.width,
                                    params.height,
                                    borderRadius,
                                    yaxisReversed,
                                    axisrotated
                                );
                            })
                            .attrs({
                                y: function(d, j) {
                                    var param = axisrotated ? 'x' : 'y';
                                    return chart.getParamValue(param, d, i, j, commonOption);
                                },
                                height: function(d, j) {
                                    var param = axisrotated ? 'width' : 'height';
                                    return chart.getParamValue(param, d, i, j, commonOption);
                                }
                            });
                    } else {
                        if (chartObj.systemConf.chart.plot.morph.enabled && chartObj.oldSeriesInfo != undefined) {
                            var morph_Obj = new zcMorph({ sampling: chartObj.systemConf.chart.plot.morph.samplingDistance });
                            var oldseriesData = chartObj.oldSeriesInfo.get(index);
                            var oldseriesIndex = oldseriesData.seriesIndex;
                            var oldseriesColor = oldseriesData.color;
                            var offsetValue = getOffsetValue(chartObj, oldseriesIndex);
                            var multicoloring = isMultiColoring(commonPlotOptions, chartObj);
                            var individualElementTransform = oldseriesData.transformationInfo;
                            transitionDuration = chartObj.systemConf.chart.plot.morph.duration;

                            var interpolated = {};
                            var line = d3_line();

                            rectEnter.each(function(d, i) {
                                var params = getFinalParams(d, i);
                                if (isNaN(params.x) || isNaN(params.y)) {
                                    params.x = -1;
                                    return null;
                                }
                                var currentPath = roundedRect(
                                    params.x,
                                    params.y,
                                    params.width,
                                    params.height,
                                    borderRadius,
                                    yaxisReversed,
                                    axisrotated
                                );
                                var fromPath = chartObj.oldSeriesInfo.get(oldseriesIndex).pathData[i];
                                var fromPathAttribute = fromPath.getAttribute('d');
                                fromPath.setAttribute('d', currentPath);
                                var toPath = fromPath.cloneNode();
                                fromPath.setAttribute('d', fromPathAttribute);
                                interpolated[i] = morph_Obj.interpolate(
                                    {
                                        path: fromPath,
                                        xOffset: offsetValue.x + individualElementTransform[i].x,
                                        yOffset: offsetValue.y + individualElementTransform[i].y
                                    },
                                    { path: toPath, xOffset: 0, yOffset: 0 }
                                );
                            });

                            rectEnter
                                .style('fill', function(d, i) {
                                    return oldseriesColor[i];
                                })
                                .transition()
                                .duration(transitionDuration)
                                .ease(easingType)
                                .style('fill', function(d, i) {
                                    if (multicoloring) {
                                        return getRawColor(chartObj, i, d, oldseriesIndex, null);
                                    } else {
                                        return getRawColor(chartObj, 0, d, oldseriesIndex, null);
                                    }
                                })
                                .attrTween('d', function(d, i) {
                                    var interpolator = interpolated[i];
                                    return function(t) {
                                        return line(
                                            interpolator.map(function(p) {
                                                return p(t);
                                            })
                                        );
                                    };
                                })
                                .on('end', function(d, i) {
                                    if (chartObj.oldSeriesInfo) {
                                        delete chartObj.oldSeriesInfo;
                                    }
                                    d3_select(this).style('fill', function() {
                                        if (multicoloring) {
                                            return fillColor(i, d, null);
                                        } else {
                                            return fillColor(0, d, null);
                                        }
                                    });
                                    rectEnter.attrs({
                                        y: function(d, j) {
                                            var param = axisrotated ? 'x' : 'y';
                                            return chart.getParamValue(param, d, i, j, commonOption);
                                        },
                                        height: function(d, j) {
                                            var param = axisrotated ? 'width' : 'height';
                                            return chart.getParamValue(param, d, i, j, commonOption);
                                        }
                                    });
                                    //});
                                    var params = getFinalParams(d, i);
                                    if (isNaN(params.x) || isNaN(params.y)) {
                                        params.x = -1;
                                        return null;
                                    }
                                    var currentPath = roundedRect(
                                        params.x,
                                        params.y,
                                        params.width,
                                        params.height,
                                        borderRadius,
                                        yaxisReversed,
                                        axisrotated
                                    );
                                    this.setAttribute('d', currentPath);
                                });
                        } else {
                            rectEnter
                                .transition()
                                .duration(isredraw ? duration : 0)
                                .delay(function(d, j) {
                                    return animationType == VERTICALALL_ANIMATION || isredraw ? 0 : (j * noofbarDatas + index) * duration;
                                })
                                .attr('d', function(d, j) {
                                    var curele = d3_select(this);
                                    var starthei = parseFloat(curele.attr('height'));
                                    endx =
                                        axisrotated && !isRangePlot
                                            ? parseFloat(curele.attr('y'))
                                            : chart.getParamValue('x', d, i, j, commonOption);
                                    endy = axisrotated
                                        ? chart.getParamValue('y', d, i, j, commonOption)
                                        : isRangePlot && !starthei
                                        ? chart.getParamValue('y', d, i, j, commonOption, null, 0)
                                        : parseFloat(curele.attr('y')); // Why (isGantt && !starthei)? -> rectEnter animation should start from y0
                                    endwid = axisrotated
                                        ? parseFloat(curele.attr('height'))
                                        : chart.getParamValue('width', d, i, j, commonOption);
                                    endhei = axisrotated ? chart.getParamValue('height', d, i, j, commonOption) : starthei;
                                    return roundedRect(endx, endy, endwid, endhei, borderRadius, yaxisReversed, axisrotated);
                                })
                                .on('end', function(d, k) {
                                    d3_select(this)
                                        .transition()
                                        .delay(function(d, j) {
                                            return animationType == VERTICALALL_ANIMATION || isredraw
                                                ? 0
                                                : (k * noofbarDatas + index) * duration;
                                        })
                                        .duration(duration)
                                        .ease(easingType)
                                        .attr('d', function(d) {
                                            var params = getFinalParams(d, k);
                                            if (isNaN(params.x) || isNaN(params.y)) {
                                                params.x = -1;
                                                return null;
                                            }
                                            return roundedRect(
                                                params.x,
                                                params.y,
                                                params.width,
                                                params.height,
                                                borderRadius,
                                                yaxisReversed,
                                                axisrotated
                                            );
                                            // return roundedRect(endx, endy, endwid, endhei, borderRadius, yaxisReversed, axisrotated);
                                        });
                                })
                                .attrs({
                                    y: function(d, j) {
                                        var param = axisrotated ? 'x' : 'y';
                                        return chart.getParamValue(param, d, i, j, commonOption);
                                    },
                                    height: function(d, j) {
                                        var param = axisrotated ? 'width' : 'height';
                                        return chart.getParamValue(param, d, i, j, commonOption);
                                    }
                                });
                        }
                    }
                } else {
                    rectEnter
                        .transition()
                        .duration(isredraw ? duration : 0)
                        .attrs({
                            d: function(d, j) {
                                var params = getFinalParams(d, j);
                                return roundedRect(
                                    params.x,
                                    params.y,
                                    params.width,
                                    params.height,
                                    borderRadius,
                                    yaxisReversed,
                                    axisrotated
                                );
                            },
                            y: function(d, j) {
                                var param = axisrotated ? 'x' : 'y';
                                return chart.getParamValue(param, d, i, j, commonOption);
                            },
                            height: function(d, j) {
                                var param = axisrotated ? 'width' : 'height';
                                return chart.getParamValue(param, d, i, j, commonOption);
                            }
                        });
                }
                if (chartObj.syncTransition) {
                    rect.exit().remove();
                } else {
                    rect.exit()
                        .transition()
                        .duration(duration)
                        .remove();
                }

                if (i > 0 && stacked) {
                    var subdividerseries = appendEle(series, 'g', ishideable ? [] : [1], '.', 'class', `subdividerseries_${i}`);
                    var dividerbarNode = subdividerseries.selectAll('rect').data(dataUpdate, function(d) {
                        return dataset.getX(d);
                    });
                    var dividerbar = dividerbarNode
                        .enter()
                        .append('rect')
                        .merge(dividerbarNode)
                        .styles({
                            fill: 'white',
                            opacity: 0.7,
                            'shape-rendering': 'crispedges'
                        })
                        .attrs({
                            width: getDividerParam('width'),
                            height: getDividerParam('height'),
                            x: function(d, j) {
                                return getDividerParam('x', d, i, j, commonOption);
                            },
                            y: function(d, j) {
                                return getDividerParam('y', d, i, j, commonOption);
                            }
                        });
                    dividerbar
                        .transition()
                        .duration(duration)
                        .ease(easingType)
                        .styles({
                            fill: 'white',
                            opacity: 0.7,
                            'shape-rendering': 'crispedges'
                        })
                        .attrs({
                            width: getDividerParam('width'),
                            height: getDividerParam('height'),
                            x: function(d, j) {
                                return getDividerParam('x', d, i, j, commonOption);
                            },
                            y: function(d, j) {
                                return getDividerParam('y', d, i, j, commonOption);
                            }
                        });
                    dividerbar.exit().remove();
                }
                if (isGantt) {
                    svg_plot_renderer.ganttHelpers.showDependencies(chart, dataUpdate, series, i);
                }

                function getDividerParam(paramType, d, i, j, cp) {
                    switch (paramType) {
                        case 'x':
                            return axisrotated ? yscale(d.y0) : chart.getParamValue('x', d, i, j, cp);
                        case 'y':
                            return axisrotated ? chart.getParamValue('y', d, i, j, cp) : yscale(d.y0);
                        case 'width':
                            return axisrotated ? 1 : barWidth;
                        case 'height':
                            return axisrotated ? barWidth : 1;
                    }
                }

                function getInitialParams(d) {
                    var v = dataset.getX(d);
                    startx = stacked ? bandwidth1(v) : bandwidth1(v) + bandwidth2(barDatasetIndex + i);
                    starty = yscale.bandwidth
                        ? yscale.origRange[0]
                        : axisrotated
                        ? stacked
                            ? hasNegValue != null
                                ? yscale(hasNegValue)
                                : yscale.range()[0]
                            : hasNegValue != null
                            ? yscale(hasNegValue)
                            : yscale.range()[0]
                        : stacked
                        ? hasNegValue != null
                            ? yscale(hasNegValue)
                            : yscale.range()[0]
                        : hasNegValue != null
                        ? yscale(hasNegValue)
                        : yscale(0) != null
                        ? mathMin(yscale(0), yscale.range()[0])
                        : yscale.range()[0];
                    starty = isWaterfall ? yscale(d.y0) : starty;
                    startwid = axisrotated ? 0 : barWidth;
                    starthei = axisrotated ? barWidth : 0;
                    return { x: startx, y: starty, width: startwid, height: starthei };
                }

                function getFinalParams(d, j) {
                    endx = chart.getParamValue('x', d, i, j, commonOption);
                    endy = chart.getParamValue('y', d, i, j, commonOption);
                    endwid = chart.getParamValue('width', d, i, j, commonOption);
                    endhei = chart.getParamValue('height', d, i, j, commonOption);
                    return { x: endx, y: endy, width: endwid, height: endhei };
                }

                chart.getMarkerPosition = isGantt ? svg_plot_renderer.ganttHelpers.getMarkerPosition : null;
            });

            if (isBoxplot) {
                var yidx = chartObj.dataset.getColumnIdx('y', yaxiscolumnorder),
                    medianidx = targetMarker.columnindex,
                    whiskeridx = whiskers.columnindex;
                var labelrenderer = chartObj.datalabels.renderer,
                    labelConfig = labelrenderer ? labelrenderer[index] : {};
                var labeltype_boxplot = [];
                if (labelConfig && labelConfig.show) {
                    var labeltype = labelConfig.labeltype;
                    var labelidx;
                    if (labeltype.constructor.name == 'Map') {
                        var indices = labeltype.get('val');
                        if (indices && indices.length == 1) {
                            //CURRENTLY MULTI CUSTOMIZED INDICES NOT SUPPORTED
                            //FOR THAT CHANGES TO BE DONE IN getCustomizedVal for parsing only particular {{val(3)}} from separator
                            labelidx = indices[0];
                            if (labelidx == medianidx) {
                                labeltype_boxplot.push(labelidx);
                            } else if (labelidx == whiskeridx || labelidx == yidx) {
                                labeltype_boxplot.push(labelidx, labelidx);
                            } else {
                                labeltype_boxplot.push('others');
                            }
                        } else {
                            labeltype_boxplot.push('others');
                        }
                    } else if (labeltype == 'y') {
                        labeltype_boxplot.push(yidx, yidx);
                    } else {
                        labeltype_boxplot.push('others');
                    }

                    labeltype_boxplot.map(function(d, i) {
                        var rangeidx = d == medianidx || d == 'others' ? null : i % 2,
                            columnidx = d == 'others' ? null : d;
                        labelQueue.push({
                            renderer: chart,
                            charttype: 'bar',
                            chartObj: chartObj,
                            rendererIndex: rendererIndex,
                            paneid: paneid,
                            args: [currentdata, index, 0, rangeidx, columnidx]
                        });
                    });
                }
            } else if (cType == 'barrange') {
                labelQueue.push({
                    renderer: chart,
                    charttype: 'bar',
                    chartObj: chartObj,
                    paneid: paneid,
                    args: [currentdata, index, 0, 1]
                });
                labelQueue.push({
                    renderer: chart,
                    charttype: 'bar',
                    chartObj: chartObj,
                    paneid: paneid,
                    args: [currentdata, index, 0, 0]
                }); //last argument in args for yrangePos
            } else {
                labelQueue.push({
                    renderer: chart,
                    charttype: 'bar',
                    chartObj: chartObj,
                    rendererIndex: rendererIndex,
                    paneid: paneid,
                    args: [currentdata, legendOrder, null, null, null, rendererIndex]
                });
            }
        });
        function getSegments(d) {
            var _data = d._data(),
                mSize = commonRendererProp.params.whiskers.size;
            if (typeof commonRendererProp.params.whiskers.size === 'string' && commonRendererProp.params.whiskers.size.indexOf('%') > -1) {
                //if user gives size as percent, '75%'
                mSize = ((parseFloat(commonRendererProp.params.whiskers.size) / 100) * commonRendererProp.params.xscale.bandwidth()) / 10;
            }
            var whiskerIndex = commonRendererProp.params.whiskers.dataindex,
                y = (yscale(d[whiskerIndex][0]) + yscale(d[whiskerIndex][1])) / 2,
                attr = commonRendererProp.params.axisrotated ? '_y' : '_x';
            var symbolSegment = getSegmentsForSymbol(
                'horizontal-whisker',
                mSize,
                _data[attr] + chart.commonOption.barWidth / 2,
                y,
                whiskerSize(d, yscale, null, null, null, whiskerIndex),
                commonRendererProp.params.axisrotated
            );
            return symbolSegment;
        }

        chart.getSegments = getSegments;
        let chartInstance = commonRendererProp.params.chartObj;
        transitionDuration = newBarParser ? parser.builderC.animationOptions._duration : transitionDuration;
        chartInstance.timerObj[chartInstance.timerObj.length] = chart.renderedTimer = setTimeout(function() {
            isredraw = true;
            chart.chartRendered = true;
        }, 2 * transitionDuration + 10);
        return chart;
    }
    chart.getHighlighter = function(chartObj, index, chartName) {
        if (!chart.highlighter) {
            var cType = chartName || getChartTypeByIndex(chartObj, null, index).name,
                isBullet = cType == 'bullet',
                isGantt = cType == 'gantt',
                isWaterfall = cType == 'waterfall',
                isBoxplot = cType == 'boxplot',
                isRadialBar = axisType == 'polar',
                whiskers = commonRendererProp.params.whiskers && commonRendererProp.params.whiskers.enabled,
                overlayWhisker = whiskers && commonRendererProp.params.whiskers.showOn == 'overlay';
            chart.highlighter = isBullet
                ? new BulletHighlighter(chartObj, chart)
                : isGantt
                ? new GanttHighlighter(chartObj, chart)
                : overlayWhisker
                ? new BarWhiskerHighlighter(chartObj, chart)
                : isBoxplot
                ? new BoxplotHighlighter(chartObj, chart)
                : isRadialBar ? new ArcHighlighter(chartObj, chart)
                : new BarHighlighter(chartObj, chart);
        }

        return chart.highlighter;
    };
    chart.commonRendererProp = function(_) {
        if (!arguments.length) {
            return commonRendererProp;
        }
        commonRendererProp = _;
        return chart;
    };

    chart.bandwidth1 = function(_) {
        if (!arguments.length) {
            return bandwidth1;
        }
        bandwidth1 = _;
        return chart;
    };

    chart.bandwidth2 = function(_) {
        if (!arguments.length) {
            return bandwidth2;
        }
        bandwidth2 = _;
        return chart;
    };

    chart.barDatasetIndex = function(_) {
        if (!arguments.length) {
            return barDatasetIndex;
        }
        barDatasetIndex = _;
        return chart;
    };

    chart.shiftInfo = function(_) {
        if (!arguments.length) {
            return shiftInfo;
        }
        shiftInfo = _;
        return chart;
    };
    chart.axisType = function(_) {
        return arguments.length ? (axisType = _, chart) : axisType;
    }
    chart.parser = function(_) {
        return arguments.length ? (parser = _, chart) : parser;
    }
    chart.destroy = function () {
        commonRendererProp.parser && commonRendererProp.parser.destroy();
    }

    return chart;
};

export var _barChart = svg_plot_renderer.bar;
