// @zohocharts/zc/src/bar/barOptions.js
import { defaultTheme as options } from '../core/defaulttheme';

options.bar = function() {
    return {
        outerPadding: 0.04,
        fillOpacity: 0.9,
        maxBandWidth: 200,
        multiColoring: false,
        border: {
            show: false,
            size: null,
            style: 'solid',
            radius: 0,
            color: null
        },
        stacked: {
            enabled: false,
            showAs: 'values' //values,percent
        },
        levelMarker: {
            enabled: false,
            color: ['#d8d8d8'],
            fillOpacity: 0.9
        },
        datalabels: {
            animation: {
                enabled: false
            },
            showAs: 'y', //x,y,percent
            stackLabels: {
                show: false,
                showAs: 'y' //y,percent
            }
        },
        animation: {
            type: 'verticalAll' //verticalAll || vertical
        },
        gradients: {
            //linear || none
            options: {
                linear: {
                    y2: 50
                }
            }
        },
        whiskers: {
            enabled: false,
            columnindex: 3,
            dataindex: 2,
            size: '50%',
            color: '#d8d8d8',
            opacity: 0.9,
            strokeWidth: 1,
            showOn: 'overlay', //'overlay'
            events: {
                enabled: true
            },
            hoveroptions: {
                strokeWidth: 2,
                opacity: 0.9,
                color: '#d8d8d8'
            }
        }
    };
};

export var _barOptions = options.bar;
