// @zohocharts/zc/src/bar/barRendererConf.js
import { DataProcessor } from '../core/data-processor';
import { mergeJSON, invokeFunction, pick } from '../utils/zcutils';
import { mathMin, mathMax } from '../core/zc-constants';
// import { getStackGroupIndex } from '../core/data-processor/transforms/stack/stack-grouping';
import { isBarChartType } from '../events/events-utils';
import { getChartTypeByIndex } from '../core/charttype-helpers';
import { stack } from '../core/main-zc-init';

export function getBarCommonOption(conf, index, legendOrder, paneid) {
    legendOrder = pick(legendOrder, index);
    var rp = conf.params,
        stacked = rp.stacked,
        chartObj = rp.chartObj,
        yscale = rp.yscale,
        processor = chartObj.dataObject,
        dataProcessorHelpers = DataProcessor.helpers,
        barDataSetIndex = chartObj.renderer[index].barDatasetIndex ? chartObj.renderer[index].barDatasetIndex() : 0,
        commonPlotOptions = rp.commonPlotOptions,
        cType_Obj = getChartTypeByIndex(chartObj, paneid, legendOrder),
        cType = cType_Obj.name,
        isBullet = cType == 'bullet',
        plotoptions = (chartObj.seriesdata[legendOrder] || chartObj.seriesdata[0]).plotoptions,
        levelMarker =
            plotoptions && plotoptions.levelMarker
                ? mergeJSON(commonPlotOptions.levelMarker, plotoptions.levelMarker, true)
                : commonPlotOptions.levelMarker,
        levelMarkerOptions = commonPlotOptions.levelMarker || {};

    var wid = chartObj.barscale2.bandwidth() || 0,
        levelMarkerWidth = wid,
        barWidth = +(commonPlotOptions.bandWidth || null) || (isBullet ? wid / 3 : wid);

    var levelMarkerBandWidth = +(levelMarkerOptions.bandWidth || wid),
        levelMarkerMaxBandWidth = +(levelMarkerOptions.maxBandWidth || wid),
        maxBandWidth = invokeFunction(commonPlotOptions.maxBandWidth, chartObj.barscale1, chartObj);
    maxBandWidth = maxBandWidth < 0 ? null : maxBandWidth;
    maxBandWidth = +(maxBandWidth || wid); //setting bandwidth if the actual bandwidth is greater than the maximum bar bandwidth

    levelMarkerWidth = mathMin(levelMarkerBandWidth, levelMarkerMaxBandWidth, wid);
    barWidth = mathMin(barWidth, maxBandWidth, wid);

    var barPadding = mathMax(0, (wid - barWidth) / 2),
        levelMarkerPadding = mathMax(0, (wid - levelMarkerWidth) / 2);

    var addRangeBand = function() {
        return yscale.bandwidth ? yscale.bandwidth() / 2 : 0;
    };

    var stackGroupIndex;
    if (stacked) {
        var allChartTypes = processor.allChartTypes,
            stackGroup = processor.getStackGroup(),
            callback = d => d.some(seriesIndex => !chartObj.seriesdata[seriesIndex].disabled && isBarChartType(allChartTypes[seriesIndex])),
            stackBarEncodes = stackGroup.filter(callback);

        stackGroupIndex = processor.categoryHierarchy ? 0 : stack.getStackGroupIndex(stackBarEncodes, index);
    } else {
        stackGroupIndex = 0;
    }

    var commonOption = {
        stacked: stacked,
        categories: rp.xscale.domain(),
        axisrotated: rp.axisrotated,
        bandwidth1: rp.xscale,
        bandwidth2: chartObj.barscale2,
        barDatasetIndex: barDataSetIndex,
        padding: barPadding,
        yscale: yscale,
        hasNegValue: rp.hasNegValue,
        yaxiscolumnorder: rp.yaxiscolumnorder,
        addRangeBand: addRangeBand,
        barWidth: barWidth,
        charttype: cType,
        levelMarker: levelMarker,
        levelMarkerWidth: levelMarkerWidth,
        levelMarkerPadding: levelMarkerPadding,
        stackGroupIndex: stackGroupIndex
    };

    chartObj.renderer[index].commonOption = commonOption;
    return commonOption;
}
