// @zohocharts/zc/src/bar/canvasBarchart.js
//$Id$
//canvasBarchart.js
import { ORDINAL_DATATYPE } from '../core/datatype-constants';
import { defined, utils, mergeJSON, toArray, isTrue, d3_values, isArray, pick } from '../utils/zcutils';
import { getFillColor, getStrokeColor, isMultiColoring, getRawColor, rgba_canvas } from '../utils/colorutils';
import { colorscale_helpers } from '../axis/colorscale-helpers';
import { getDashArray } from '../utils/dashstyles';
import { mathMin, deg2rad } from '../core/zc-constants';
// import { legend_helpers } from '../legend/legend-helper';
import { getBarCommonOption } from './barRendererConf';
import { canvasUtils } from '../utils/canvasutils';
import { charttype, legend_helpers, plot, canvas_plot_renderer, stack } from '../core/main-zc-init';
import { effects } from '../effects/zceffects';
// import { getStackGroupIndex } from '../core/data-processor/transforms/stack/stack-grouping';
import { BulletHighlighter } from '../bullet/bullet-highlight';
import { BarHighlighter } from './bar-highlight';
import { GanttHighlighter } from '../gantt/gantt-highlight';
import { BoxplotHighlighter } from '../boxplot/boxplot-highlight';
import { DataProcessor } from '../core/data-processor';
import { getChartTypeByIndex } from '../core/charttype-helpers';
import { getImageByIndex } from '../helpers/image-pallete-helpers';
import { isFacet } from '../axis/hierarchy-axis-utils';
import { isDataActive } from '../core/data-processor/is-data-active';

canvas_plot_renderer.bar = function() {
    var commonRendererProp,
        bandwidth1,
        bandwidth2,
        yscale,
        barDatasetIndex,
        shiftInfo,
        isredraw = false;

    function chart(context, seriesdata, effect, filterIndex) {
        var rp = commonRendererProp.params,
            chartObj = rp.chartObj,
            xscale = rp.xscale,
            stacked = rp.stacked,
            index = rp.datasetIndex,
            axisrotated = rp.axisrotated,
            commonPlotOptions = rp.commonPlotOptions,
            color = rp.color,
            paneColor = rp.paneColor,
            labelQueue = rp.labelQueue,
            hasNegValue = rp.hasNegValue,
            yaxiscolumnorder = rp.yaxiscolumnorder,
            dataset = rp.dataset,
            legendOrder = rp.legendOrderIndex,
            rendererIndex = rp.rendererIndex,
            systemconf = chartObj.systemConf,
            plotarea = chartObj.plotarea,
            dataObject = chartObj.dataObject,
            dataProcessorHelpers = DataProcessor.helpers,
            isSelectionContainer = !(chartObj.container.select('.selection-canvas').empty()),
            isSelectionCanvas = context.canvas.className == 'selection-canvas' ? true : false,
            selectionArray = chartObj.cache.clickSelectionData;

        // categories = xscale.domain();
        // categories = chartObj.dataObject.getCategories("x");
        seriesdata = chartObj.dataObject.categoryHierarchy ? seriesdata[index] : chartObj.seriesdata[index];

        var paneid = isFacet(chartObj) ? `${seriesdata.row}_${seriesdata.column}` : '0_0',
            cType_Obj = getChartTypeByIndex(chartObj, paneid, index);
        if (!bandwidth1 || !bandwidth2) {
            //Test_515.json (Combinational chart(bar + line) where all the bar series are disabled)
            isredraw = true;
            chart.chartRendered = true;
            return;
        }

        yscale = rp.yscale;
        var datatypeY = chartObj.dataset.getDataType('y', yaxiscolumnorder),
            xCategories = xscale.domain(),
            yCategories = datatypeY == ORDINAL_DATATYPE ? yscale.domain() : null,
            categories = defined(dataObject.hierarchical) ? dataObject.getCategories('x') : xCategories;

        var cType = cType_Obj.name;
        var gradientDim = null; //[w,h];
        var yAxisReversed = (rp.yaxisreversed = chartObj.systemConf.chart.axes.yaxis[yaxiscolumnorder || 0].reversed),
            xAxisReversed = (rp.xaxisreversed = chartObj.systemConf.chart.axes.xaxis.reversed);

        var plotoptions = seriesdata.plotoptions;
        var gvp = (rp.gvp = utils.getValueOfPlotOptions(plotoptions, commonPlotOptions)),
            isBullet = cType_Obj.value === charttype.bullet,
            isBarrange = cType_Obj.value === charttype.barrange;

        var targetMarker = (fillOpacity = gvp('targetMarker') || {});

        var levelMarker =
                plotoptions && plotoptions.levelMarker
                    ? mergeJSON(commonPlotOptions.levelMarker, plotoptions.levelMarker, true)
                    : commonPlotOptions.levelMarker,
            borderOptions =
                plotoptions && plotoptions.border
                    ? mergeJSON(commonPlotOptions.border, plotoptions.border, true)
                    : commonPlotOptions.border;
        var multiColoring = isMultiColoring(commonPlotOptions, chartObj),
            hasColorScale = colorscale_helpers.isQuantileColorScale(chartObj);

        var fillStyle = (chart.fillColorFunc = getFillColor(chartObj, legendOrder, paneid, gradientDim, null, null, null, rp)),
            strokeColor = getStrokeColor(chartObj, index, paneid, null, rp);

        var fillOpacity = gvp('fillOpacity'),
            getParamValue = (chart.getParamValue = function() {
                return chartObj.dataObject.getTransRectValue.apply(chartObj.dataObject, toArray(arguments));
            });

        var commonOption = getBarCommonOption(chartObj.rendererConf[rendererIndex], rendererIndex, legendOrder);
        var levelMarkerWidth = commonOption.levelMarkerWidth,
            levelMarkerPadding = commonOption.levelMarkerPadding;
        levelMarker.overlayRange = 'below';
        levelMarker.hasBaseValue = levelMarker.hasBaseValue ?? isBarrange;
        var levelMarkerColor = levelMarker.color; //Saving the levelMarker colors, helps in multi categories bullet if each categories dont have same no of levels

        var borderRadius = {
            tlx: 0,
            tly: 0,
            trx: 0,
            trY: 0,
            brx: 0,
            bry: 0,
            blx: 0,
            bly: 0
        };
        var borderSize = 0,
            borderColor = null,
            borderStyle = null,
            radius = 0;
        if (defined(borderOptions) && isTrue(borderOptions.show)) {
            borderSize = defined(borderOptions.size) ? borderOptions.size : 1;
            borderColor = borderOptions.color;
            borderStyle = borderOptions.style;
            radius = borderOptions.radius;
            borderRadius = getParamValue('radius', null, null, null, commonOption, radius);
        }

        var dashStyle = getDashArray(borderStyle, borderSize == null ? 2 : borderSize);

        var legend_data = systemconf.legend;
        var rounded = isRounded(borderRadius),
            patternName = legend_data.patternPallete,
            patternPallete = legend_data.patternPallete;

        var size = pick(targetMarker.size, levelMarkerWidth - 2 * pick(targetMarker.padding, commonOption.padding / 2)),
            bandwidth = isBullet ? xscale.bandwidth() : levelMarkerWidth,
            customMarkerSize = targetMarker.size;

        targetMarker.size = size = defined(customMarkerSize) ? (isBullet ? mathMin(size, bandwidth / 10) : size) : size / 10;

        const zc_gradient = effect && effect.gradient;
        defined(effect) ? (effect = effect.pattern || effect) : null;

        //STACKED BAR : for which series the level marker should be drawn
        var isLevelMarkerIndex;
        if (stacked && levelMarker.enabled && !defined(effect)) {
            var actualStackGroupIndex = stack.getStackGroupIndex(dataObject.getStackGroup(), index),
                stackEncode = chartObj.dataObject.getEncode('stack', actualStackGroupIndex);

            isLevelMarkerIndex = stackEncode && stackEncode.seriesIndices[0] === index;
        }

        if (!seriesdata.disabled) {
            chart.seriesdataAfterUpdate = [];
            seriesdata.data.forEach(function(data, i) {
                var dataUpdate = chartObj.dataObject.updateSeriesdatabyCategories(
                    d3_values(data),
                    xCategories,
                    yCategories,
                    yaxiscolumnorder,
                    true
                );

                if (hasColorScale && defined(filterIndex)) {
                    dataUpdate = legend_helpers.filterSelectionByColor(dataUpdate, true, filterIndex, chartObj);
                }
                if (chartObj.renderVisiblePointsOnly) {
                    dataUpdate = dataProcessorHelpers.filterPointsInViewport(chartObj, dataUpdate, 'bar', 'x', index, i);
                }
                chart.seriesdataAfterUpdate[i] = dataUpdate;

                if (chartObj.systemConf.chart.axes.xaxis.reversed) {
                    dataUpdate = dataUpdate.slice(0).reverse();
                }

                if (stacked && levelMarker.enabled && !defined(effect) && xCategories.length != dataUpdate.length) {
                    plot.render.bulletHelpers.addDummyDataForLevelmarker(dataUpdate, xCategories, dataset, yaxiscolumnorder);
                }

                if(isSelectionContainer){
                    if(defined(selectionArray[0])){
                        chartObj.container.select('.zc-canvas').style("visibility", "hidden");
                    }
                    else{
                        chartObj.container.select('.selection-canvas').style("visibility", "hidden");
                        chartObj.container.select('.zc-canvas').style("visibility", "visible");
                    }
                }

                dataUpdate.forEach(function(d, j) {
                    var x,
                        y,
                        w,
                        h,
                        gradient,
                        fillColor = getRawColor(chartObj, multiColoring ? j : legendOrder, d, index, paneid),
                        selectedElement = defined(d.clickSelection) && d.clickSelection,
                        isDataInActive =  !isDataActive(d, chartObj, index);

                    if (levelMarker.enabled && !defined(effect)) {
                        if ((stacked && isLevelMarkerIndex && i === 0) || !stacked) {
                            var rect = plot.render.bulletHelpers.ZC_BAR_getLevelMarkerRect(
                                    d,
                                    levelMarker,
                                    chartObj,
                                    commonOption,
                                    rp,
                                    i,
                                    index
                                ),
                                x = axisrotated ? rect.y : rect.x,
                                y = axisrotated ? rect.x : rect.y,
                                w = rect.width,
                                h = rect.height;

                            context.save();
                            context.beginPath();
                            context.fillStyle =
                                plot.render.bulletHelpers.getBulletLevelMarkerGradient(
                                    d,
                                    levelMarker,
                                    chartObj,
                                    fillColor,
                                    levelMarkerColor
                                ) ||
                                rgba_canvas(
                                    (isArray(levelMarker.color) ? levelMarker.color[0] : levelMarker.color) || fillColor,
                                    levelMarker.fillOpacity
                                );
                            canvasUtils.roundedRect(context, x, y, w, h, borderRadius, yAxisReversed, axisrotated);
                            context.fill();
                            context.closePath();
                            context.restore();
                        }
                    }

                    x = getParamValue('x', d, i, j, commonOption);
                    y = getParamValue('y', d, i, j, commonOption);
                    w = getParamValue('width', d, i, j, commonOption);
                    h = getParamValue('height', d, i, j, commonOption);

                    if (yAxisReversed) {
                        if (axisrotated) {
                            x = x - w;
                        } else {
                            y = y - h;
                        }
                    }
                    gradientDim = [w, h];

                    var drawBarFunction = function(context, fillStyle) {
                        context.save();
                        context.translate(x, y);
                        context.strokeStyle = rgba_canvas(fillColor, fillOpacity);
                        context.fillStyle = fillStyle;
                        if (rounded) {
                            context.beginPath();
                            canvasUtils.roundedRect(context, 0, 0, w, h, borderRadius);
                            context.fill();
                        } else {
                            context.beginPath();
                            context.rect(0, 0, w, h);
                            context.fill();
                        }
                        setBorder(context, borderSize, borderColor || fillColor, dashStyle);
                        context.closePath();
                        context.restore();
                    };

                    if (chartObj.imagePallete) {
                        let imageUrl = getImageByIndex(chartObj, multiColoring ? j : legendOrder, paneid);
                        effects.canvas_patterns.imageFill({ x: x, y: y, context: context, name: imageUrl }, drawBarFunction);
                    } else {
                        var fillStyle = defined(effect)
                            ? effect
                            : chart.fillColorFunc(
                                  multiColoring ? j : legendOrder,
                                  d,
                                  null,
                                  paneColor,
                                  fillOpacity,
                                  gradientDim,
                                  { x: x, y: y, width: w, height: h, context: context },
                                  false
                              );

                        if(isSelectionCanvas){
                            fillStyle =  selectedElement || isDataInActive ?
                            chart.fillColorFunc(
                                multiColoring ? j : legendOrder,
                                d,
                                null,
                                paneColor,
                                fillOpacity,
                                gradientDim,
                                { x: x, y: y, width: w, height: h, context: context },
                                false
                            )
                            : effect;
                        }

                        zc_gradient ? drawBarFunction(context, zc_gradient) : null;
                        drawBarFunction(context, fillStyle);
                    }

                    // draw target marker
                    if (isBullet && targetMarker.enabled === true && !d.isDummy) {
                        x = bandwidth1(dataset.getX(d)) + bandwidth / 2;
                        y = yscale(d[targetMarker.dataindex]);
                        drawTargetMarker(context, targetMarker, size, x, y, axisrotated);
                    }
                });
            });
        }
        targetMarker.size = customMarkerSize;
        labelQueue.push({
            renderer: chart,
            charttype: 'bar',
            chartObj: chartObj,
            rendererIndex: rendererIndex,
            paneid,
            args: [seriesdata, legendOrder, null, null, null, rendererIndex]
        });

        chart.chartRendered = true;
        return chart;
    }
    chart.getHighlighter = function(chartObj, index, chartName) {
        if (chart.highlighter) {
            return chart.highlighter;
        }
        var cType = chartName || getChartTypeByIndex(chartObj, null, index).name,
            isBullet = cType == 'bullet',
            isGantt = cType == 'gantt',
            isWaterfall = cType == 'waterfall',
            isBoxplot = cType == 'boxplot';
        chart.highlighter = isBullet
            ? new BulletHighlighter(chartObj, chart)
            : isGantt
            ? new GanttHighlighter(chartObj, chart)
            : isBoxplot
            ? new BoxplotHighlighter(chartObj, chart)
            : new BarHighlighter(chartObj, chart);
        return chart.highlighter;
    };

    chart.commonRendererProp = function(_) {
        if (!arguments.length) {
            return commonRendererProp;
        }
        commonRendererProp = _;
        return chart;
    };
    chart.bandwidth1 = function(_) {
        if (!arguments.length) {
            return bandwidth1;
        }
        bandwidth1 = _;
        return chart;
    };
    chart.bandwidth2 = function(_) {
        if (!arguments.length) {
            return bandwidth2;
        }
        bandwidth2 = _;
        return chart;
    };
    chart.barDatasetIndex = function(_) {
        if (!arguments.length) {
            return barDatasetIndex;
        }
        barDatasetIndex = _;
        return chart;
    };

    chart.shiftInfo = function(_) {
        if (!arguments.length) {
            return shiftInfo;
        }
        shiftInfo = _;
        return chart;
    };

    function isRounded(radius) {
        for (var l in radius) {
            if (radius.hasOwnProperty(l)) {
                return !!radius[l];
            }
        }
        return false;
    }

    /**
     *
     * @param context
     * @param marker
     * @param size
     * @param x
     * @param y
     * @param axisRotated
     */

    function drawTargetMarker(context, marker, size, x, y, axisRotated) {
        var outerSize = size * 5,
            outerStrokeWidth = pick(marker.outerStrokeWidth, 1),
            innerStrokeWidth = pick(marker.innerStrokeWidth, 1),
            outerStrokeColor = marker.outerStrokeColor,
            innerStrokeColor = marker.innerStrokeColor,
            outerFillColor = marker.outerFillColor,
            innerFillColor = marker.innerFillColor;

        if (defined(marker.outerFillOpacity) && outerFillColor) {
            outerFillColor = rgba_canvas(outerFillColor, marker.outerFillOpacity);
        }
        if (defined(marker.outerStrokeOpacity) && outerStrokeColor) {
            outerStrokeColor = rgba_canvas(outerStrokeColor, marker.outerStrokeOpacity);
        }
        if (defined(marker.innerFillOpacity) && innerFillColor) {
            innerFillColor = rgba_canvas(innerFillColor, marker.innerFillOpacity);
        }
        if (defined(marker.innerStrokeOpacity) && innerStrokeColor) {
            innerStrokeColor = rgba_canvas(innerStrokeColor, marker.innerStrokeOpacity);
        }

        // fix the dash style
        if (defined(marker.dashStyle)) {
            marker.outerDashStyle = marker.innerDashStyle = marker.dashStyle;
        }

        // using scatter marker
        var cache = canvas_plot_renderer.scatter.getCache(
            marker.symbol,
            outerSize,
            outerStrokeColor,
            outerStrokeWidth,
            outerFillColor,
            innerStrokeColor,
            innerStrokeWidth,
            innerFillColor,
            marker
        );
        var adjust = cache.adjustPosition,
            cacheCanvas = cache.canvas,
            drawWidth = cache.width,
            centered = outerSize + outerStrokeWidth / 2 + adjust;

        context.save();
        // rotate by 90 degree if the axis rotated
        if (axisRotated) {
            context.translate(y, x);
            context.rotate(deg2rad * 90); // rotate 90 deg
        } else {
            context.translate(x, y);
        }

        context.drawImage(cacheCanvas, -centered, -centered, drawWidth, drawWidth);
        context.restore();
    }

    function setBorder(context, size, color, dashStyle) {
        if (size > 0) {
            context.lineWidth = size;
            if (isArray(dashStyle) && dashStyle.length && defined(context.setLineDash)) {
                context.setLineDash(dashStyle);
            }
            context.strokeStyle = color;
            context.stroke();
        }
    }
    return chart;
};

export var _canvasBarChart = canvas_plot_renderer.bar;
