// @zohocharts/zc/src/bar/dotGrid.js
//$Id$
//dotGrid.js
import { svg_plot_renderer } from '../core/main-zc-init';
import { appendEle, wrapTextContent } from '../utils/zcsvgutils';
import { TRANSPARENT, NONE } from '../core/zc-constants';
import { effects } from '../effects/zceffects';
import { bbox } from '../utils/zcutils';
import { d3_select } from '../core/d3-constants';
/*
 *10X10 grid, where percentage value is shown in highlighted dots
 */
svg_plot_renderer.dotgrid = function() {
    var commonRendererProp;
    function chart(selection, data) {
        var rp = commonRendererProp.params,
            chartObj = rp.chartObj,
            xscale = rp.xscale,
            stacked = rp.stacked,
            index = rp.datasetIndex,
            axisrotated = rp.axisrotated,
            commonPlotOptions = rp.commonPlotOptions,
            color = rp.color,
            labelQueue = rp.labelQueue,
            hasNegValue = rp.hasNegValue,
            yaxiscolumnorder = rp.yaxiscolumnorder,
            dataset = rp.dataset,
            dataObject = chartObj.dataObject,
            systemconf = chartObj.systemConf,
            plotarea = chartObj.plotarea;

        var currentdata = chartObj.seriesdata[0];
        var plotoptions = currentdata.plotoptions;
        var currentseriesdata = currentdata.data[0];

        var container = (g = d3_select(selection[0][0]));
        var categories = dataObject.getCategories('x');
        var bestLayout = getBestLayout(plotarea, categories.length);

        var startX = 0,
            startY = 0;
        var rows = bestLayout.rows,
            columns = bestLayout.columns;
        var margin = 10;
        var dotGridSize = 10;
        var minSize = Math.min(bestLayout.calcWid, bestLayout.calcHei);
        dotGridSize = minSize > margin * 2 ? minSize - margin * 2 : 0;

        /*
         * Dot grid series
         */
        var dotGridSeries = appendEle(g, 'g', [1], '.', 'class', 'dotgridseries');
        var cellsGroup = appendEle(dotGridSeries, 'g', currentseriesdata, '.', 'class', 'cells');

        /*
         * Background rect prop
         */

        // var backgroundRect = appendEle(cellsGroup, "rect", [1], ".", "class", "background");
        // dotGridSeries.selectAll("rect.background").attrs({
        //     "x":function(d,i){
        //         return getX(i,0);
        //     },
        //     "y":function(d,i){
        //         return getY(i,0);
        //     },
        //     "width":bestLayout.actWid,
        //     "height":bestLayout.actHei
        // }).styles({
        //     "stroke":NONE,
        //     "fill":NONE
        // });

        /*
         * Dot grid rect prop
         */

        var dotGridRectLayer_1 = appendEle(cellsGroup, 'rect', [1], '.', 'class', 'dotGridLayer_1').attr('id', 'layers');
        var dotGridRectLayer_2 = appendEle(cellsGroup, 'rect', [1], '.', 'class', 'dotGridLayer_2').attr('id', 'layers');
        dotGridSeries.selectAll('rect#layers').attrs({
            width: dotGridSize,
            height: dotGridSize
        });
        dotGridSeries
            .selectAll('rect.dotGridLayer_1')
            .attrs({
                x: function(d, i) {
                    return getX(i, margin);
                },
                y: function(d, i) {
                    return getY(i, margin);
                },
                mask: function(d, i) {
                    var prop = { x: getX(i, margin), y: getY(i, margin), size: dotGridSize, fillColor: 'white' };
                    return gridFill(prop, dataset.getY(currentseriesdata[i], 0), 'top2Bottom');
                }
            })
            .styles({
                stroke: TRANSPARENT,
                fill: '#ececec'
            });
        dotGridSeries
            .selectAll('rect.dotGridLayer_2')
            .attrs({
                x: function(d, i) {
                    return getX(i, margin);
                },
                y: function(d, i) {
                    return getY(i, margin);
                },
                mask: function(d, i) {
                    var prop = { x: getX(i, margin), y: getY(i, margin), size: dotGridSize, fillColor: rp.color[0] };
                    return gridFill(prop, dataset.getY(currentseriesdata[i], 0), 'bottom2Top');
                }
            })
            .styles({
                stroke: TRANSPARENT,
                fill: rp.color[0]
            });

        /*
         * Text content
         */
        var textWid = bestLayout.actWid;
        var textHei = 15; //bestLayout.actHei - (dotGridSize+margin);
        var percentageTextMargin = textHei + margin; //(bestLayout.actHei - (dotGridSize+margin))/2;
        var fontSize = textHei - percentageTextMargin;
        var textY = textHei;

        var textG = appendEle(cellsGroup, 'text', [currentseriesdata], '.', 'class', 'grid');
        dotGridSeries.selectAll('text.grid').each(function(d, i) {
            var label = d3_select(this);
            wrapTextContent(label, [d[i][0], d[i][1]], { lineHeight: 0 });
        });
        dotGridSeries
            .selectAll('text.grid')
            .attrs({
                transform: function(d, i) {
                    var bb = bbox(this);
                    var transX = getX(i, 0) + bestLayout.calcWid / 2,
                        transY = getY(i, margin) + percentageTextMargin + dotGridSize;
                    return `translate(${transX},${transY})`;
                }

                // "x":function(d,i){
                //     var bb = this.getBBox();
                //     var widthTransform = textWid / bb.width;
                //     var heightTransform = textHei / bb.height;
                //     var value = widthTransform < heightTransform ? widthTransform : heightTransform;
                //     return (getX(i,0)+(bestLayout.calcWid/2));
                // },
                // "y":function(d,i){
                //     var bb = this.getBBox();
                //     var widthTransform = textWid / bb.width;
                //     var heightTransform = textHei / bb.height;
                //     var value = widthTransform < heightTransform ? widthTransform : heightTransform;
                //     return (getY(i,margin)+percentageTextMargin+dotGridSize);
                // }

                // "transform":function(d,i){
                //     var bb = this.getBBox();
                //     var widthTransform = textWid / bb.width;
                //     var heightTransform = textHei / bb.height;
                //     var value = widthTransform < heightTransform ? widthTransform : heightTransform;
                //     return "matrix("+value+", 0, 0, "+value+", 0,0)";
                // }
            })
            .styles({
                stroke: NONE,
                fill: 'black',
                'text-anchor': 'middle',
                'font-size': '13px'
            });
        alignLayout('center', dotGridSeries, plotarea);
        function getX(i, padding) {
            var colIdx = i % columns;
            return startX + padding + colIdx * bestLayout.actWid;
        }
        function getY(i, padding) {
            var rowIdx = parseInt(i / columns) % rows;
            return startY + padding + rowIdx * bestLayout.actHei;
        }
        return chart;
    }
    chart.commonRendererProp = function(_) {
        if (!arguments.length) {
            return commonRendererProp;
        }
        commonRendererProp = _;
        return chart;
    };
    function drawBackgroundRect() {}
    function drawDotGridRect() {}
    function drawText() {}

    function gridFill(options, percent, type) {
        var cellSize = options.size / 10;
        var patternProp = {
            size: cellSize.toString(),
            bgClr: options.fillColor,
            x: options.x,
            y: options.y,
            type: 'circle',
            stroke: TRANSPARENT,
            strokeWidth: '1',
            fillOpacity: '1'
        };
        var patternURL = `url(#${effects.get(effects.svg_patterns, 'svgPatternFill', patternProp)})`;
        var maskProp = {
            size: cellSize,
            width: options.size,
            height: options.size,
            x: options.x,
            y: options.y,
            fillColor: patternURL,
            percent: percent,
            type: type
        };
        var maskURL = `url(#${effects.get(effects.svg_masks, 'tenbytenGrid', maskProp)})`;
        return maskURL;
    }

    function getBestLayout(viewport, count) {
        var bestLayout = {
            rows: 0,
            columns: 0,
            totalArea: 0
        };
        for (var i = 1; i <= count; i++) {
            for (var j = 1; j <= count; j++) {
                if (i * j >= count) {
                    var currentArea = calculateArea(viewport, count, i, j);
                    if (currentArea.totalArea > bestLayout.totalArea) {
                        bestLayout = currentArea;
                    }
                }
            }
        }
        return bestLayout;
    }
    function calculateArea(viewport, count, rows, columns) {
        var optimizedGrid = getGridSize({
            width: viewport.width / columns,
            height: viewport.height / rows
        });
        var totalArea = optimizedGrid.width * optimizedGrid.height * count;
        return {
            rows: rows,
            columns: columns,
            actWid: viewport.width / columns,
            actHei: viewport.height / rows,
            calcWid: optimizedGrid.width,
            calcHei: optimizedGrid.height,
            totalArea: totalArea
        };
    }
    function getGridSize(viewport) {
        // Cross-multiplications:
        //     600            : 800
        //     viewport.width : height
        //     width          : viewport.height
        if ((viewport.width * 800) / 600 <= viewport.height) {
            return {
                width: viewport.width,
                height: (viewport.width * 800) / 600
            };
        }
        return {
            width: (viewport.height * 600) / 800,
            height: viewport.height
        };
    }
    function alignLayout(mode, node, viewport) {
        var chartArea = bbox(node);
        var transX = 0,
            transY = 0;
        if (mode == 'center') {
            if (chartArea.height < viewport.height) {
                var centerPos = viewport.height / 2;
                transY = centerPos - chartArea.height / 2;
            }
            if (chartArea.width < viewport.width) {
                var centerPos = viewport.width / 2;
                transX = centerPos - chartArea.width / 2;
            }
            node.attr('transform', `translate(${transX},${transY})`);
        }
    }
    return chart;
};

export var _dotgridChart = svg_plot_renderer.dotgrid;
