// @zohocharts/zc/src/bar/markerPointer.js
//$Id$
//markerPointer.js
import { addSVGmarker } from '../utils/svg-symbols';
import { appendEle } from '../utils/zcsvgutils';
import { mathMin } from '../core/zc-constants';
import { d3_select } from '../core/d3-constants';

export function addMarkerPointer(group, data, comparatorData, comparisonConfig, config) {
    var chartObj = config.chart,
        isAnimationEnable = config.animationOptions.enabled,
        axisrotated = config.axisrotated,
        renderer = config.renderer,
        seriesindex = config.seriesindex,
        barSize = axisrotated
            ? renderer.getParamValue('height', data[0], seriesindex, null, renderer.commonOption)
            : renderer.getParamValue('width', data[0], seriesindex, null, renderer.commonOption),
        mSize = mathMin(barSize / 10, 1),
        markerPosition = getMarkerCoordinates(comparatorData),
        symbolSize = {
            inner: 0,
            outer: 1
        },
        symbolType = axisrotated ? 'triangle-left' : 'triangle-down';
    var markerG = appendEle(group, 'g', comparisonConfig.enabled ? [1] : [], '.', 'class', 'zcmarkerPointers');

    var marker = appendEle(markerG, 'g', markerPosition, '.', 'class', 'markers', null, function(d, i) {
        // #ZC2049
        return comparatorData[i] && comparatorData[i].y;
    });

    marker.each(function(d, j) {
        var m = d3_select(this);
        m.attrs({
            transform: function(d) {
                return `translate(${d[0]},${d[1]})`;
            }
        })
            .style('opacity', 0)
            .transition()
            .duration(isAnimationEnable ? config.animationOptions.duration : 0)
            .style('opacity', 1);
        addSVGmarker(
            m,
            null,
            {
                size: mSize,
                symbol: symbolType,
                outerFillOpacity: comparisonConfig.strokeOpacity,
                outerStrokeOpacity: comparisonConfig.strokeOpacity,
                outerStrokeWidth: 0
            },
            {
                symbolSize: symbolSize,
                color: comparisonConfig.strokeColor || 'black'
            },
            chartObj
        );
    });
}

function getMarkerCoordinates(comparatorData) {
    var positions = comparatorData.map(function(d, i) {
        return comparatorData[i][3];
    });
    return positions;
}
