// @zohocharts/zc/src/scale/evaluators/correction/marker-evaluator.js
//$Id$
import { mathMax } from '../../../core/zc-constants';
import { d3_extent, d3_merge } from '../../../core/d3-constants';
import { invokeFunction, splat, isObject } from '../../../utils/zcutils';
import { isFunction } from '../../../utils/function-utils';
import { ScaleRegistry } from '../../../scale/scale-registry';
import { bubbleCorrection } from '../../evaluators/correction/correction-adapter'; // TODO: Remove at end of Code modularization
import { _modularized } from '../../../core/zcconf';

export function evalMarkerCorrection(_) {
    let _scale = this,
        { data, size, position, scale } = _;
    // TODO: check this
    scale = invokeFunction(scale, _scale) || (d => d);
    let domain = invokeFunction(scale.domain),
        bandwidth = invokeFunction(scale.bandwidth) || 0,
        range = isFunction(position) ? [scale(domain[0]), scale(domain[domain.length - 1]) + bandwidth] : [0, 0],
        bbs = MeasureMarker(data, { size, position, scale, _this: _scale });

    let bound = d3_extent(d3_merge([d3_extent(bbs, bb => bb.x), d3_extent(bbs, bb => +(bb.x + bb.width).toFixed(12)), [0]]));

    if (!_modularized.scaleCorrection) {
        let c = bubbleCorrection(scale, bound, _scale);
        if (c) {
            return c;
        }
    }

    return [mathMax(range[0] - bound[0], 0), mathMax(bound[1] - range[1], 0)];
}

function MeasureMarker(data, _args) {
    let { scale, size, position, _this } = _args,
        bandwidth = invokeFunction(scale.bandwidth) || 0;
    return data.map(d => {
        let s = splat(size).reduce((prev, cur) => prev + invokeFunction(cur, d), 0),
            _position = invokeFunction(true, position, _this, d, scale),
            p = isObject(_position) ? _position.value : scale(_position) + bandwidth / 2; // TODO: enhance

        return {
            x: p - s / 2,
            y: p - s / 2,
            width: s,
            height: s
        };
    });
}

ScaleRegistry.add('marker-correction-evaluator', evalMarkerCorrection);
