// @zohocharts/zc/src/plot/redrawTools.js
import { setDefaultOptions } from '../core/defaulttheme';
import { setPropVal } from '../utils/zcutils';
import { chartcategory } from '../core/main-zc-init';

export function updatePlotOption(chartObj, newSystemConf) {
    var plotoptions = newSystemConf.chart.plot.plotoptions;
    for (var i = 0; i < chartObj.distinctChartTypes.length; i++) {
        setDefaultOptions(chartObj.distinctChartTypes[i], plotoptions, chartObj.globalProp);
    }
}

export function updateExistingPlotComponents(chartObj) {
    var isSingleSourceEventCharts = chartcategory.singleSourceEvent;

    var nonAxis = chartcategory.nonaxis;
    var nonAxisMultiSeriesCategory = chartcategory.nonaxismultiseries;

    var nonAxisCharts = nonAxis.concat(nonAxisMultiSeriesCategory);
    var polarAxis = chartcategory.polarAxis;
    var oldchart = chartObj.oldChartTypes;

    chartObj.chartTypes.map((d, i) => {
        if (
            (!(nonAxisCharts.indexOf(oldchart[i].value) != -1) && nonAxisCharts.indexOf(d.value) != -1) ||
            polarAxis.indexOf(oldchart[i].value) != -1
        ) {
            removeAxis(chartObj);
            removeAxisLabels(chartObj, i);
            removeThreshold(chartObj);
            removePlotAreaBorder(chartObj);
        }
        if (isSingleSourceEventCharts.indexOf(d.value) === -1) {
            chartObj.eventHandler.removeTracker();
        }
        removeDataLabels(chartObj);
    });
}

function removeAxis(chartObj) {
    var svg = chartObj.svg;
    svg.selectAll('.x-axis')
        .selectAll('g')
        .remove();
    svg.selectAll('.y-axis')
        .selectAll('g')
        .remove();
}

function removeAxisLabels(chartObj, i) {
    setPropVal(chartObj.axes, 'x.axisLabel', null);
    setPropVal(chartObj.axes, 'y.groups[' + i + '].axisLabel', null);
}

function removeThreshold(chartObj) {
    var staticPane = chartObj.staticPaneSVG;
    staticPane
        .selectAll('.threshold')
        .selectAll('g')
        .remove();
}

function removePlotAreaBorder(chartObj) {
    var staticPane = chartObj.staticPaneSVG;
    staticPane
        .selectAll('.plotareaBorder-property')
        .selectAll('line')
        .remove();
}

function removeDataLabels(chartObj) {
    chartObj.staticPaneSVG
        .selectAll('.labelholder')
        .selectAll('g')
        .remove();
}

export function checkChartTypeEquality(fromChartTypes, toChartTypes) {
    var result = true;
    if (fromChartTypes === undefined) {
        return result;
    }
    if (fromChartTypes.length != toChartTypes.length) {
        return false;
    }
    fromChartTypes.map((d, i) => {
        if (toChartTypes[i].name != d.name) {
            result = false;
        }
    });
    return result;
}
