// @zohocharts/zc/src/plot/renderer.js
//$Id$
//renderer.js
import { Registry } from '../core/zc-registry';
import { plot, labelRenderer, charttypenames, charttype, axis, overview } from '../core/main-zc-init';
import { chartcategory, renderer_helpers } from '../core/main-zc-init';
import { stateChangeListener } from '../events/eventActions';
import { isDocumentVisible, simpleClone, arrayIncludes, d3_values, defined, pick, getDistinct, arrayFind, invokeFunction } from '../utils/zcutils';
import { isObjectPropDefined } from '../utils/zcutils';
import { isFunction } from '../utils/function-utils';
import { DataProcessor } from '../core/data-processor';
import { isBarChartType, isHeatMapChartType, isWaterfallChartType } from '../events/events-utils';
import { ORDINAL_DATATYPE } from '../core/datatype-constants';
import { VizLayout } from '../core/main-constructor';
import { $Browser } from '../core/browser-detect';
import { getLeafAxisOrder, isFacet } from '../axis/hierarchy-axis-utils';
import { getAxisData } from '../axis/axisutils';
import { getChartSeriesColor } from '../core/seriescolor-helpers';
import { hasColorColumnInNonAxis } from '../utils/colorutils';
import { getAllSeriesTypes } from '../core/series-parser-helpers';
import { getGridList } from '../core/data-processor/pivot-layout-helpers';
import { getChartTypeByIndex } from '../core/charttype-helpers';
import { getDynamicSeriesPosition } from './addRemoveSeries/addRemoveSeriesUtilities';
import { renderDom } from '../mark/dom-render';
// import { duplicateOverview } from '../behaviours/visual-effects/chart/controllers/overview';
import { d3_event, d3_selectAll } from '../core/d3-constants';
import { reStyleDom } from '../mark/dom-paint';
// import { PlotHighlighter } from '../events/visualEffects/highlight/plot-highlighter';
import { clearselectionContext, clearHighlightContext } from '../events/visualEffects/highlight/highlight-context';
import { ZC_ChartEvents } from '../events/chart-events';

export * from '../plot/rendererConf';

plot.render = function(chartObj) {
    var mode = chartObj.systemConf.chart.plot.renderer.mode.toLowerCase();

    var render = {
            mode: hasSvgOnly(chartObj) ? 'svg' : mode.toLocaleLowerCase()
        },
        args = {};

    args.labelQueue = [];
    args.notesArray = [];
    args.labelArray = {};

    if (!chartObj.rendererConf) {
        chartObj.rendererConf = [];
    }

    if (typeof plot[`${render.mode}Renderer`] === 'function') {
        render[render.mode] = plot[`${render.mode}Renderer`](chartObj, args);
    } else {
        throw new SyntaxError(`${render.mode}Renderer not found`);
    }

    render.draw = function(update) {
        chartObj.chartRendered = false;
        args.labelArray = {};
        args.labelQueue = [];
        if (chartObj.annotation) {
             chartObj.annotation.events = [];
        } else { 
            chartObj.annotation = { events: [] }; // TODO: check
        }
        this[this.mode].draw(update);
        
        let _annotationDelay = annotationDelay(chartObj);
        if (_annotationDelay !== 'chartRendered') {
            chartObj.timerObj[chartObj.timerObj.length] = setTimeout(function() {
                renderAnnotation(chartObj, args);
            }, _annotationDelay);
        }
    };
    return render;
};

renderer_helpers.renderCompletedEvents = function(chartObj, update, args, callback) {
    var systemconf = chartObj.systemConf,
        isDialChart = chartObj.dataObject.hasLevelMarker && !chartObj.isAxisCategory,
        DoWaitForChartRendered = wait4Rendered(chartObj),
        renderer = chartObj.renderer,
        k,
        axesRendered = false,
        isSelection = defined(chartObj.cache.clickSelectionData),
        isSelectionArray = chartObj.cache.clickSelectionData;
    // var labelHolder = chartObj.staticPaneSVG.select(".labelholder").style("opacity", 0);

    if (chartObj._isAxisCategory() || chartObj.dataObject.isPolarAxisCategory) {
        chartObj.timerObj[chartObj.timerObj.length] = setTimeout(function() {
            axesRendered = true;
        }, chartObj.axes.animationDuration + 17);
    } else {
        axesRendered = true;
    }

    var checkRendered = function() {
        if (isDocumentVisible()) {
            chartObj.chartRendered = true;
        }

        if (chartObj.destroying) {
            clearInterval(timer);
            return true;
        }

        k = renderer.length;
        while (k--) {
            if (renderer[k].chartRendered !== true || !axesRendered || (chartObj.racing  && !chartObj.racing.ended && chartObj.racing.preventEvent) ) {
                return false;
            }
            // args.notesArray = renderer[k].commonRendererProp().notes();
        }
        if(chartObj.customVisual){
            chartObj.CustomVisualPromise.then(d => {
                chartObj._customVisualRendered = true;
                chartObj.chartRendered = true;
                // return true;
            })
            if(chartObj._customVisualRendered !== true){
                return false;
            }
        }
        proceed();

        clearInterval(timer);

        if (isFunction(callback)) {
            callback();
        } else {
            // ZC1740
            chartObj.updating = false;
            chartObj.chartRendered = true;
        }

        return true;
    };

    function proceed() {
        try {
            // labelHolder.transition().duration(250).style("opacity", 1);

            // fixes #ZC1767 (ticks hiding on chrome while scrolling)
            if (chartObj.divScroll && $Browser.CHROME) {
                chartObj.divScroll.fixScrollView(true, chartObj.canvasarea, chartObj);
            }
            // if(chartObj.viewFlow){
            //     chartObj.viewFlow.bound = {
            //         canvasarea: chartObj.canvasarea, 
            //         chartarea: chartObj.chartarea, 
            //         plotarea:chartObj.plotarea
            //     }
            // }
            // if(chartObj.viewFlow && chartObj.viewFlow.scales){
            //     for(let id in chartObj.viewFlow.scales){
            //         if(!chartObj.viewFlow.scales[id].component){
            //             let _parserConf = chartObj.viewFlow.scales[id].conf,
            //             parser_expression_val = applyExpressionValueForParser(_parserConf, chartObj.viewFlow);

            //             parser_expression_val.map(d => {
            //                 _parserConf[d[0]] = d[1];
            //             })

            //             var newScaleComp1 = new Scale(null, 0, _parserConf, chartObj);
            //             let newScaleObj = newScaleComp1.getScale();
            //             chartObj.viewFlow.scales[id].component = newScaleObj;
            //         }
            //     }
            // }
            // if(chartObj.viewFlow && chartObj.viewFlow.marks._custom.length){
            //     chartObj.viewFlow.marks._custom.map(mark => {
            //         let parser = invokeFunction(mark.mark.parser, chartObj);
            //         renderDom(parser, null, chartObj.viewFlow);
            //     })
                
            // }

            if(chartObj.viewFlow && chartObj.viewFlow.triggerQueue.length){
                chartObj.viewFlow.triggerQueue.map(obj => {
                    let { trigger, react } = obj;
                    let ele = d3_selectAll(trigger.element);
                    trigger.on.map(t => {
                        ele.on(t.event, function(event, d){
                            d3_event(event);
                            
                            if(t.value == 'self'){
                                react.value = this.value;
                            }else if(t.value == "datum"){
                                react.value = this.__data__;
                            }else if(typeof t.value === 'function'){
                                react.value = invokeFunction(t.value, this.__data__);
                            }else{
                                react.value = react.value || 1;
                            }
                            react.value = t.value == 'self' ? this.value : t.value == "datum" ? this.__data__ : react.value;

                            if(t.singleNodeDataFinder) {
                                var eventHandler = chartObj.eventHandler,
                                    node = eventHandler.getSourceNode();
                                var mouse = eventHandler.getRelativeMousePosition(node, true);
                                let nearest_data = ZC_ChartEvents.findData.call(chartObj.eventHandler.eventManager, mouse);
                                react.value = nearest_data.nearest;
                            }
                    
                            //handle the dependencies
                            react.dependencies.map(d => {
                                if(d.recalculate){
                                    if(d.component == "scale"){
                                        let dependencies_scale = chartObj.viewFlow.scales[d.id].dependencies;
                                        dependencies_scale.map(module => {
                                            if(module.type == 'axis'){
                                                
                                                chartObj.plotarea = simpleClone(chartObj.___chartarea)
                                                chartObj.plotarea_approx = null;
                                                axis.render(chartObj, chartObj.chartWidth, chartObj.chartHeight, true);
                                            }else if(module.type == 'mark'){
                                                renderDom(module.parser, module.parent, chartObj.viewFlow);
                                            }
                                        })
                                    }else{
                                        renderDom(d.parser, null, chartObj.viewFlow);
                                    }
                                    
                                }else{
                                    reStyleDom((d.parser.getElement()), {config: d.config}, {__tokens__:chartObj.viewFlow});
                                }
                                
                            })
                            
                        })
                    })
                })
            }
            stateChangeListener(chartObj, systemconf.chart.plot.events, 'plot', update ? 'onreload' : 'onload');
            callDatalableAndAnnotationRenderer(chartObj, args, !(DoWaitForChartRendered && update), update);
            chartObj.handleHiddenContainer(true);
            if(chartObj.overview){
                overview.duplicateOverview && overview.duplicateOverview(chartObj);
            }
        
            chartObj.timerObj[chartObj.timerObj.length] = setTimeout(function() {
                // setTimeout is needed for Analytics - Dial chart export as HTML issue
                stateChangeListener(chartObj, systemconf.canvas.events, 'chart', update ? 'onreload' : 'onload');
            }, 30);
        } catch (err) {
            VizLayout.helpers.handleErrors(chartObj, err);
        }

        // reset the chart phase
        chartObj.phase = null;
    }

    var timer = (chartObj.timerObj[chartObj.timerObj.length] = setInterval(checkRendered, 1000 / 60));

    if (DoWaitForChartRendered) {
        callDatalableAndAnnotationRenderer(chartObj, args, isDialChart || update, update);
    }

    if(isSelection && defined(isSelectionArray[0])){
        selectionUpdate(this.mode, chartObj);
    }

};
export function getChartName(chartName) {
    switch (chartName) {
        case 'waterfall':
        case 'gantt':
        case 'boxplot':
        case 'bullet':
        case 'barrange': {
            chartName = 'bar';
            break;
        }
        case 'web':
        case 'arearange': {
            chartName = 'area';
            break;
        }
        case 'bubble':
        case 'geobubble':
        case 'geoscatter': {
            chartName = 'scatter';
            break;
        }
        case 'bubblepie':
        case 'geobubblepie': {
            chartName = 'bubblepie';
            break;
        }
    }
    return chartName;
}

function callDatalableAndAnnotationRenderer(chartObj, args, enable, update) {
    var systemConf = chartObj.systemConf,
        dataLabels = systemConf.chart.plot.datalabels;
    if (chartObj.systemConf.__$overview) {
        return false;
    }
    /*
     * It will call datalabels renderer and annotation renderer
     */
    if (enable) {
        if (labelRenderer && args.labelQueue.length && !dataLabels.renderOnInit) {
            //MODULARIZE
            var datalabelEvents = dataLabels.events;

            stateChangeListener(chartObj, datalabelEvents, 'plot', 'init');

            let dataLabelComponent = Registry.getComponent('labelRenderer');
            dataLabelComponent(args.labelQueue, args.labelArray, chartObj);
            stateChangeListener(chartObj, datalabelEvents, 'plot', update ? 'onreload' : 'onload');
        }
        args.labelArray = {};
        args.labelQueue = [];
        if (annotationDelay(chartObj) === 'chartRendered') {
            if (chartObj.dataObject.isFlowChartCategory) {
                renderAnnotation(chartObj, args);
            } else {
                chartObj.timerObj[chartObj.timerObj.length] = setTimeout(function() {
                    renderAnnotation(chartObj, args);
                }, 17);
            }
        }
    }
}

function renderAnnotation(chartObj, args) {
    let notes = chartObj.systemConf.notes,
        notesEnabled = defined(notes) && notes.enabled === true,
        notesArray;

    if (notesEnabled) {
        let annotation = Registry.getComponent('annotation');
        if (!annotation._empty) {
            notesArray = annotation.draw(chartObj, args.notesArray);

            // ZC3288 - on redraw it needs some time
            setTimeout(function() {
                pick(isObjectPropDefined(chartObj, 'annotation.flatten', true), []).forEach(_ => _());
            }, 0);
        }
    } else {
        chartObj.notes.selectAll('*').remove();
        chartObj.container.selectAll('g.zc-annotation').remove(); // annotations with custom parent
    }

    args.notesArray = notesArray || [];
}

function annotationDelay(chartObj) {
    return pick(isObjectPropDefined(chartObj.systemConf, 'notes.animation.delay', true), 'chartRendered');
}

function wait4Rendered(chartObj) {
    let dataObject = chartObj.dataObject;

    if (isFacet(chartObj)) {
        let allDistinctSeriesTypes_val = chartObj.distinctChartTypes || getDistinct(getAllSeriesTypes(chartObj.userdata)),
            allDistinctSeriesTypes = allDistinctSeriesTypes_val.map(_ => charttype[_]);
        return allDistinctSeriesTypes.every(chartType => arrayIncludes(chartcategory.axis, chartType));
    }

    let isDialChart = dataObject.hasLevelMarker && !dataObject.isAxisCategory;
    return (
        ((dataObject.isAxisCategory || dataObject.isPolarAxisCategory) && !dataObject.pieWithAxis) ||
        isDialChart ||
        (chartObj.isGeoMap && !dataObject.hasBubblePie)
    );
}

renderer_helpers.arrangeSeriesRenderingOrder = function(chartObj, data, startIdx, facets) {
    var seriesdata = data || chartObj.seriesdata,
        dataProcessorHelpers = DataProcessor.helpers;

    if (chartObj.dataObject.isCombinationalChart) {
        var order = {
                defaultType: 0,
                area: 0,
                arearange: 0,
                geoheatmap: 0,
                bar: 1,
                barrange: 1,
                line: 2,
                bubble: 3,
                geobubble: 3,
                bubblepie: 4,
                geobubblepie: 4,
                scatter: 5,
                geoscatter: 5
            },
            globalChartType = chartObj.userdata.seriesdata.type;

        if (facets) {
            seriesdata = data.map(d => {
                return d.node;
            });
        }
        var seriesdata_cpy = simpleClone(seriesdata);
        var counter = startIdx || 0;
        seriesdata_cpy.map(function(d, i) {
            d.counter = i;
            d.seriesidx = counter++;
            return d;
        });
        seriesdata_cpy.sort(function(a, b) {
            var type1 = dataProcessorHelpers.getChartType(a.type || globalChartType, 'string'),
                type2 = dataProcessorHelpers.getChartType(b.type || globalChartType, 'string');
            var order1 = order[type1] == null ? order.defaultType : order[type1],
                order2 = order[type2] == null ? order.defaultType : order[type2],
                seriesidx1 = a.seriesidx,
                seriesidx2 = b.seriesidx;
            return order1 < order2
                ? -1
                : order1 > order2
                ? 1 //Arrange by chart type
                : seriesidx1 < seriesidx2
                ? -1
                : seriesidx1 > seriesidx2
                ? 1
                : 0; //Arrange by series index
        });
        if (facets) {
            return data.map((d, i) => {
                d.node = seriesdata[seriesdata_cpy[i].counter];
                return d;
            });
        }
        return seriesdata_cpy;
    }
    return seriesdata;
};

renderer_helpers.setRendererProp = function(chartObj, renderer, seriesdata, index, renderingOrder, args, rendererIdx, charttype, layerIdx) {
    rendererIdx = pick(rendererIdx, index);
    //index - legendIndex -
    //renderingOrder - sorted seriesdata for combochart
    //rendererIdx - for grids refers overall serial counter
    var cTypeObj = chartObj.chartTypes[index];
    var cType = charttype ? charttypenames.get(charttype) : cTypeObj.name;
    var cTypeNumeric = charttype || cTypeObj.value;
    var systemconf = chartObj.systemConf,
        helpers = DataProcessor.helpers;
    var hierarchical = chartObj.dataObject.hierarchical,
        facetRendering = hierarchical && hierarchical.facetRendering,
        rowList,
        columnList,
        rowCount,
        clustersCount,
        hasMultiCluster,
        hasMultiLayers,
        columnLeavesCount,
        layers;

    if (facetRendering) {
        (rowList = hierarchical.pivotLayout.getRootNodeList('row')),
            (columnList = hierarchical.pivotLayout.getRootNodeList('column')),
            (rowCount = rowList.getCount()),
            (clustersCount = rowList.getClustersCount()),
            (hasMultiCluster = rowList.hasMultiCluster()),
            (columnLeavesCount = columnList.getLeavesCount()),
            (layers = hierarchical.pivotLayout.layers.groups[layerIdx]),
            (hasMultiLayers = hasMultiCluster ? hierarchical.pivotLayout.layers.isLayerNode : false);
        var gridOptions = hierarchical.gridOptions,
            paneid = `${seriesdata[index].row}_${seriesdata[index].column}`;
        if (gridOptions[paneid]) {
            gridOptions[paneid].renderer[index] = renderer;
        }
    }
    var legendIdx = facetRendering ? layers.indexOf(seriesdata[index].value) : index;
    setCommonRendererProp(renderer, seriesdata, index, cType, cTypeNumeric, legendIdx, rendererIdx);
    renderer.commonRendererProp(chartObj.rendererConf[rendererIdx]);
    var yaxiscolumnorder = hasMultiLayers
        ? seriesdata[legendIdx].parent.node.yaxiscolumnorder
        : chartObj.seriesdata[index].yaxiscolumnorder;
    var data = hasMultiLayers ? seriesdata[legendIdx] : chartObj.seriesdata[legendIdx];
    var yscindex = helpers.getParsedYAxisOrder(yaxiscolumnorder);

    renderer.chartRendered = false;

    if (chartObj._isAxisCategory() || chartObj.dataObject.isPolarAxisCategory) {
        var yaxis = chartObj.axes.y.groups[yscindex];
        var ysc = yaxis.scale || chartObj.axes.y.groups[0].scale;

        var yscidx = facetRendering && (hasMultiCluster || rowCount) ? seriesdata[index].row || 0 : null,
            xscale = facetRendering
                ? columnLeavesCount > 1
                    ? chartObj.axes.x.leaves[seriesdata[index].column].scale
                    : chartObj.axes.x.leafScaleGrid
                : chartObj.axes.x.scale,
            yscale = facetRendering && (hasMultiCluster || rowCount) ? chartObj.axes.y.leaves[yscidx].scale : ysc;
        var dynamicSeriesPosition = chartObj.renderer[index].commonRendererProp().params.dynamicSeriesPosition;
        if (chartObj.seriesAddRemoveDetails && chartObj.seriesAddRemoveDetails.seriesAddRemovePhase) {
            dynamicSeriesPosition = getDynamicSeriesPosition(chartObj, index);
        }

        chartObj.rendererConf[rendererIdx]
            .dynamicSeriesPosition(dynamicSeriesPosition)
            .xscale(xscale)
            .yscale(yscale)
            .shapescale(defined(chartObj.dataset.shapeidx) ? arrayFind(chartObj.scales, _ => _.id === 'shape-0').getScale() : null)
            .datasetIndex(index)
            .renderingOrderIndex(renderingOrder)
            .legendOrderIndex(index)
            .legendIndex(legendIdx) // TODO: check legendOrderIndex, datasetIndex
            .legendOrder(layerIdx) // TODO: check legendOrderIndex, datasetIndex
            .rendererIndex(rendererIdx)
            .axisrotated(chartObj.axes.rotated)
            .hasNegValue(chartObj.dataObject.containsNegativeData(yscindex))
            .stacked(chartObj.dataObject.isstacked(cTypeNumeric));

        args.labelArray = chartObj.rendererConf[rendererIdx].params.labelArray;
        args.labelQueue = chartObj.rendererConf[rendererIdx].params.labelQueue;
    }

    if (chartObj.dataObject.isFlowChartCategory) {
        chartObj.rendererConf[rendererIdx].datasetIndex(index);
    }

    if (isBarChartType(cTypeNumeric) && !isHeatMapChartType(cTypeNumeric)) {
        renderer.bandwidth1(chartObj.barscale1).bandwidth2(chartObj.barscale2);
        args.barDatasetIndex = facetRendering && chartObj.barscale2.domain().length > 1 ? legendIdx : args.barDatasetIndex; //TODO for pivot with subseries data , pivot2 case
        if (!args.image) {
            renderer.barDatasetIndex(args.barDatasetIndex);
        }
        if (data.disabled == null || !data.disabled) {
            args.barDatasetIndex = args.barDatasetIndex + data.data.length;
        }
    }

    if (isWaterfallChartType(cTypeNumeric)) {
        var commonParams = chartObj.rendererConf[rendererIdx].params;
        commonParams.categoryColorConfig = renderer_helpers.getWaterfallCategoriesConfig(
            chartObj.seriesdata[index].plotoptions,
            commonParams.commonPlotOptions
        );
    }

    chartObj.rendererConf[rendererIdx].bound(renderer_helpers.getBound(chartObj, seriesdata[index]));

    function setCommonRendererProp(renderer, seriesdata, index, cType, actualCType, legendIdx, rendererIdx) {
        var hierarchical = chartObj.dataObject.hierarchical,
            facetRendering = hierarchical && hierarchical.facetRendering,
            paneColor = facetRendering && seriesdata[index] ? seriesdata[index].node.color : null,
            hasMultiCluster = facetRendering && hierarchical.pivotLayout.rows && hierarchical.pivotLayout.rows.hasMultiCluster(),
            hasMultiLayers = hasMultiCluster ? hierarchical.pivotLayout.layers.isLayerNode : false;

        var currentseriesdata = facetRendering && hasMultiLayers ? seriesdata[legendIdx].node : chartObj.seriesdata[legendIdx];
        var paneid = seriesdata[index] ? `${seriesdata[index].row}_${seriesdata[index].column}` : '0_0';
        var seriesClr = facetRendering
            ? getChartSeriesColor(chartObj, paneid, index)
            : currentseriesdata.color || getChartSeriesColor(chartObj, paneid, index); //previously it is legendIdx; Now values are properly updated in gridoptions
        var yaxiscolumnorder = hasMultiLayers
            ? seriesdata[legendIdx].parent.node.yaxiscolumnorder
            : chartObj.seriesdata[index].yaxiscolumnorder;
        var yscindex =
            facetRendering && hierarchical.hasRows()
                ? getLeafAxisOrder(seriesdata[index].row || 0, null, 'y', chartObj)
                : helpers.getParsedYAxisOrder(yaxiscolumnorder);
        var commonProp = chartObj.rendererConf[rendererIdx];
        commonProp
            .chartObj(chartObj)
            .dataset(chartObj.dataset)
            .yaxiscolumnorder(yscindex)
            .commonPlotOptions(chartObj.systemConf.chart.plot.plotoptions[cType])
            .color(
                cType == 'bubblepie'
                    ? chartObj.seriesColor
                    : chartObj.isAxisCategory
                    ? checkNgetColorOverlayData(renderer, seriesdata, seriesClr, index, yscindex, cType, rendererIdx)
                    : chartObj.dataObject.isPolarAxisCategory || hasColorColumnInNonAxis(chartObj)
                    ? seriesClr
                    : chartObj.seriesColor
            )
            .paneColor(paneColor)
            .notes(args.notesArray)
            .updateLabelArrayAndLabelQueue(args.labelArray, args.labelQueue);
    }

    function checkNgetColorOverlayData(renderer, seriesdata, seriesClr, index, yscindex, cType, rendererIdx) {
        var xaxis_userdata = isObjectPropDefined(chartObj.userdata, 'chart.axes.xaxis') ? chartObj.userdata.chart.axes.xaxis : {},
            yaxes_userdata = isObjectPropDefined(chartObj.userdata, 'chart.axes.yaxis') ? chartObj.userdata.chart.axes.yaxis : {},
            yaxes_threshold = yaxes_userdata != null && yaxes_userdata[yscindex] != null && yaxes_userdata[yscindex].threshold != null ? yaxes_userdata[yscindex].threshold  : null,
            xaxis_threshold = isObjectPropDefined(xaxis_userdata, 'threshold') ? xaxis_userdata.threshold : null;
        var thOverlay_data_x = xaxis_threshold ? xaxis_threshold.coloroverlay || {} : {};
        var thOverlay_data_y = yaxes_threshold ? yaxes_threshold.coloroverlay || {} : {};
        var showColorOverlay = thOverlay_data_x.show || (thOverlay_data_y != null && thOverlay_data_y.show);
        if (!showColorOverlay) {
            return seriesClr;
        } else {
            var type = thOverlay_data_y != null && thOverlay_data_y.show ? 'y' : thOverlay_data_x.show ? 'x' : null,
                overlayData =
                    thOverlay_data_y != null && thOverlay_data_y.show ? thOverlay_data_y : thOverlay_data_x.show ? thOverlay_data_x : null;
            var negThreshType = thOverlay_data_y != null && thOverlay_data_y.show ? 'y' : thOverlay_data_x.show ? 'x' : null;
            var colorOverlay = showColorOverlay ? axis.threshold.overlay(chartObj, 'continous') : null,
                axisOrder = negThreshType === 'y' ? yscindex : 0;

            chartObj.rendererConf[rendererIdx]
                .threshold(showColorOverlay ? overlayData : null)
                .thresholdType(negThreshType)
                .colorOverlay(showColorOverlay ? axis.threshold.overlay(chartObj, 'discrete', type, overlayData) : null);

            return cType == 'bar' || cType == 'scatter' || cType == 'bubble'
                ? seriesClr
                : colorOverlay(seriesClr, overlayData, axisOrder, negThreshType);
        }
    }
};

renderer_helpers.getBound = function(chartObj, gridData) {
    let { hierarchical } = chartObj.dataObject,
        facetRendering = hierarchical && hierarchical.facetRendering;

    if (facetRendering) {
        let { row, column } = gridData,
            { pivotLayout } = hierarchical,
            columnList = pivotLayout.columns,
            rowList = pivotLayout.rows,
            columnSize = columnList.leafSize,
            rowSize = rowList.leafSize,
            //
            firstColumnBranchCount = columnList.count ? columnList.branches[0].domain.length : 1,
            firstRowBranchCount = rowList.count ? rowList.branches[0].domain.length : 1,
            //
            columnCount = firstColumnBranchCount * columnList.clustersCount,
            rowCount = firstRowBranchCount * rowList.clustersCount,
            //
            columnBranchCount = columnList.branches.length,
            rowBranchCount = rowList.branches.length,
            //
            x,
            y,
            width,
            height,
            polar = chartObj.cache.polar && chartObj.cache.polar[`${row}-${column}`],
            xReversed = columnBranchCount && getAxisData(chartObj, 'x', 0).reversed,
            yReversed = rowBranchCount && getAxisData(chartObj, 'y', 0).reversed;

        if (chartObj.axes.rotated) {
            x = (yReversed ? rowCount - row - 1 : row) * rowSize;
            y = (xReversed ? column : columnCount - column - 1) * columnSize;
            width = rowSize;
            height = columnSize;
        } else {
            x = (xReversed ? columnCount - column - 1 : column) * columnSize;
            y = (yReversed ? rowCount - row - 1 : row) * rowSize;
            width = columnSize;
            height = rowSize;
        }

        return { width: width || 0, height: height || 0, x: x || 0, y: y || 0, polar };
    } else {
        let { width, height } = chartObj.plotarea,
            polar = chartObj.plotarea_approx;
        return { width, height, polar, x: 0, y: 0, left: 0, top: 0, polar };
    }
};

renderer_helpers.saveSeriesdataInRenderer = function(chartObj, seriesIndex, rendererIndex) {
    rendererIndex = pick(rendererIndex, seriesIndex);
    var rp = chartObj.rendererConf[seriesIndex].params, // TODO: check the rendererConf for pivot
        xscale = rp.xscale,
        yscale = rp.yscale,
        yaxiscolumnorder = rp.yaxiscolumnorder,
        dataset = chartObj.dataset,
        datatypeX = dataset.getDataType('x'),
        datatypeY = dataset.getDataType('y', yaxiscolumnorder),
        currentseriesdata = chartObj.seriesdata[seriesIndex].data,
        renderer = chartObj.renderer[rendererIndex];

    if (datatypeX == ORDINAL_DATATYPE || datatypeY == ORDINAL_DATATYPE) {
        renderer.seriesdataAfterUpdate = [];
        var dataProcessorHelpers = DataProcessor.helpers,
            isCatNullX = dataProcessorHelpers.isCatNull(dataset, 'x'),
            isCatNullY = dataProcessorHelpers.isCatNull(dataset, 'y', yaxiscolumnorder),
            xCategories = datatypeX == ORDINAL_DATATYPE ? xscale.domain() : null,
            yCategories = datatypeY == ORDINAL_DATATYPE ? yscale.domain() : null;
        currentseriesdata.forEach(function(d, i) {
            renderer.seriesdataAfterUpdate[i] = chartObj.dataObject.updateSeriesdatabyCategories(
                d3_values(d),
                xCategories,
                yCategories,
                yaxiscolumnorder
            );
        });
    } else {
        renderer.seriesdataAfterUpdate = null;
    }
};

function hasSvgOnly(chartObj) {
    if (isFacet(chartObj)) {
        var gridList = getGridList(chartObj.dataObject);
        return gridList.some(grid => {
            let paneID = `${grid.row}_${grid.column}`;
            var cType = getChartTypeByIndex(chartObj, paneID, 0).value;
            return arrayIncludes(chartcategory.svgOnly, cType);
        });
    } else {
        var cType = chartObj.chartTypes[0];
        return chartcategory.svgOnly.indexOf(cType.value) > -1 || chartObj.dataObject.pieWithAxis;
    }
}

function selectionUpdate(mode, chart){// To maintain selection after redraw
    var legendConfig = chart.systemConf.legend,
        legendAPIs = chart.legend.component;

    chart.chartTypes.forEach(function(chartType){
        var ctype = chartType['name'],
        indexLength = chart.isAxisCategory || chartObj.isAxisCategory == undefined ? chart.seriesdata.length : chart.seriesdata[0].data[0].length,
        indexLength = ctype === 'sankey' ? chartObj.cache.sankey.legend.length : indexLength,
        indexArray = Array(indexLength).fill().map((_, i) => i),
        indexArray = ctype === 'sankey' ? nodeBasedArray(chart, indexArray) : indexArray;
        var plotHighlighter = chart.eventHandler.plotHighlighter || new (Registry.getComponent('highlighter.PlotHighlighter'))(chart);

            function nodeBasedArray(chartObj, array){
                array.forEach(function(ii){
                    if(chartObj.cache.disabledItems['$node'+ii]){
                        let removeIndex = chartObj.cache.sankey.legendItemsMap['$node'+ii]['index'];
                        array = array.filter(function(i){
                            return i !== removeIndex;
                        });
                    }
                });
            return array; 
            }
            if(ctype == "scatter" || ctype == "line" || ctype == "area"){
                chartObj.cache.selection.plotHighlightedPoints = defined(chartObj.cache.selection.plotHighlightedPoints) ?  chartObj.cache.selection.plotHighlightedPoints : []; 
                
                chartObj.cache.selection.plotHighlightedPoints.forEach(function(d, i){
                        var className = `highlight${d.paneid}${d.seriesIndex}${d.itemIndex}${undefined}`;
                        clearselectionContext(chartObj, className);
                        clearHighlightContext(chartObj, className);
                        if(d.clickSelection && !chartObj.seriesdata[d.seriesIndex].disabled){
                            plotHighlighter.highlight(d, true);
                        }
                });
            }
            
            legendAPIs.highlightSeries(null, chart.seriesdata, indexArray, legendConfig, chart, false ,true);
    }); 
}
