// @zohocharts/zc/src/plot/addRemoveSeries/addSeries.js
import { chartcategory, charttypenames, charttype, renderer_helpers } from '../../core/main-zc-init';
import { getDistinct, isUndefined, splat } from '../../utils/zcutils';
import { DataProcessor } from '../../core/data-processor';
import { GROUP_SELECTOR } from '../morph/seriesGroupMapping';
import { d3_map } from '../../core/d3-constants';
import { getChartName } from '../renderer';
import { createElement } from '../../utils/zcsvgutils';
// import { ReverseLegendHighlight } from '../../events/reverse-legend-highlight';
import { getUniqueChartTypes } from '../../core/charttype-helpers';
import { VizLayout } from '../../core/main-constructor';
import { updateAddRemoveSeriesProperty } from './addRemoveSeriesUtilities';
import { Registry } from '../../core/zc-registry';

var reConstructChartData = function(seriesInfo, seriesIndices, chart) {
    var chartData = chart.userdata.seriesdata.chartdata,
        oldSeriesDataLength = chartData.length,
        seriesAddedAtEnd = true,
        seriesDataIndexMap = d3_map(),
        length;
    if (seriesIndices.length === 1 && isUndefined(seriesIndices[0])) {
        seriesIndices.splice(0, 1);
    }
    length = oldSeriesDataLength + seriesIndices.length;

    //mapping of input seriesinfo wih input seriesindex
    seriesInfo.map((d, i) => {
        if (isUndefined(seriesIndices[i])) {
            seriesIndices[i] = length;
            length++;
        }
        seriesDataIndexMap.set(seriesIndices[i], d);
    });
    //sorting of seriesIndices
    seriesIndices.sort(function(a, b) {
        return a - b;
    });
    if (chart.nochartdata) {
        chart.rendererConf = [];
    }
    seriesIndices.map(i => {
        var index = i,
            d = seriesDataIndexMap.get(i);
        chart.seriesAddRemoveDetails.maxDOMseriesIndex++;
        chart.seriesAddRemoveDetails.seriesPositionMap.set(index, chart.seriesAddRemoveDetails.maxDOMseriesIndex);
        chartData.splice(index, 0, d);
        chart.renderer.splice(index, 0, null);
        chart.rendererConf.splice(index, 0, null);
        if (i < oldSeriesDataLength) {
            seriesAddedAtEnd = false;
        }
    });

    chart.staticPaneSVG
        .selectAll('.qualitativeGroup')
        .selectAll('g')
        .remove();
    //notes removal
    chartObj.notes.selectAll('*').remove();
    chartObj.container.selectAll('g.zc-annotation').remove(); 

    chart.seriesAddRemoveDetails.seriesUpdatedAtEnd = seriesAddedAtEnd;
    return seriesDataIndexMap;
};

function getClassInfoForAddSeries(newSeries, oldSeries, chart) {
    //old series is the series before which the new series will get added
    var newSeriesType = getChartName(newSeries.type),
        newSeriesClass = [
            `.${newSeriesType}${GROUP_SELECTOR.get(newSeriesType)}${chart.seriesAddRemoveDetails.seriesPositionMap.get(newSeries.index)}`
        ],
        oldSeriesClass;

    if (!isUndefined(oldSeries.data)) {
        var oldSeriesType = getChartName(oldSeries.type),
            dynamicPosition;
        if (chart.renderer[oldSeries.index] === null) {
            dynamicPosition = chart.seriesAddRemoveDetails.seriesPositionMap.get(oldSeries.index);
        } else {
            if (
                chartObj.renderer[oldSeries.index] &&
                chartObj.renderer[oldSeries.index].commonRendererProp() &&
                chartObj.renderer[oldSeries.index].commonRendererProp().params &&
                !isUndefined(chartObj.renderer[oldSeries.index].commonRendererProp().params.dynamicSeriesPosition)
            ) {
                dynamicPosition = chart.renderer[oldSeries.index].commonRendererProp().params.dynamicSeriesPosition;
            } else {
                dynamicPosition = chart.renderer[oldSeries.index].commonRendererProp().params.datasetIndex;
            }
        }
        oldSeriesClass = [`.${oldSeriesType}${GROUP_SELECTOR.get(oldSeriesType)}${dynamicPosition}`];
    }

    if (newSeriesType === 'line') {
        newSeriesClass.push(`.scatterseries_${chart.seriesAddRemoveDetails.seriesPositionMap.get(newSeries.index)}`);
        if (oldSeriesClass) {
            oldSeriesClass.push(`.scatterseries_${dynamicPosition}`);
        }
    }
    if (newSeriesType === 'area') {
        newSeriesClass.push(`.lineseries_${chart.seriesAddRemoveDetails.seriesPositionMap.get(newSeries.index)}`);
        newSeriesClass.push(`.scatterseries_${chart.seriesAddRemoveDetails.seriesPositionMap.get(newSeries.index)}`);
        if (oldSeriesClass) {
            oldSeriesClass.push(`.lineseries_${dynamicPosition}`);
            oldSeriesClass.push(`.scatterseries_${dynamicPosition}`);
        }
    }
    return {
        newSeriesClass: newSeriesClass,
        oldSeriesClass: oldSeriesClass
    };
}
function insertSeriesToDOM(chart, seriesIndices) {
    var isComboChart = (chart.dataObject.isCombinationalChart = chart.dataObject.pivot
            ? getUniqueChartTypes(chart).values().length > 1
            : getDistinct(DataProcessor.helpers.getAllChartTypes(chart.systemConf, 'object'), d => charttypenames.get(d)).length > 1),
        seriesholder = chart.svg.selectAll('.seriesholder');

    if (isComboChart) {
        var orderedSeriesdata = isComboChart ? renderer_helpers.arrangeSeriesRenderingOrder(chart) : chart.seriesdata;
        for (var i = orderedSeriesdata.length - 1, ele; i >= 0; i--) {
            var currData = orderedSeriesdata[i],
                currSeriesIdx = currData.seriesidx;
            if (chart.renderer[currSeriesIdx] === null) {
                if (orderedSeriesdata[i + 1]) {
                    ele = getClassInfoForAddSeries(
                        {
                            data: currData,
                            index: currSeriesIdx,
                            type: DataProcessor.helpers.getChartType(currData.type, 'string', currSeriesIdx, chart.userdata)
                        },
                        {
                            data: orderedSeriesdata[i + 1],
                            index: orderedSeriesdata[i + 1].seriesidx,
                            type: DataProcessor.helpers.getChartType(
                                orderedSeriesdata[i + 1].type,
                                'string',
                                orderedSeriesdata[i + 1].seriesidx,
                                chart.userdata
                            )
                        },
                        chart
                    );
                } else {
                    ele = getClassInfoForAddSeries(
                        {
                            data: currData,
                            index: currSeriesIdx,
                            type: DataProcessor.helpers.getChartType(currData.type, 'string', currSeriesIdx, chart.userdata)
                        },
                        {},
                        chart
                    );
                }
                if (!ele.oldSeriesClass) {
                    ele.oldSeriesClass = [];
                }
                ele.newSeriesClass.map((d, i) => {
                    createElement('g', d, [{ data: currData, index: i }], seriesholder, ele.oldSeriesClass[i], d => d);
                });
            }
        }
    } else {
        for (var i = seriesIndices.length - 1, ele; i >= 0; i--) {
            var currData = chart.seriesdata[i],
                currSeriesIdx = i;
            if (chart.seriesdata[i + 1]) {
                ele = getClassInfoForAddSeries(
                    {
                        data: chart.seriesdata[i],
                        index: seriesIndices[i],
                        type: DataProcessor.helpers.getChartType(chart.seriesdata[i].type, 'string', seriesIndices[i], chart.userdata)
                    },
                    {
                        data: chart.seriesdata[seriesIndices[i] + 1],
                        index: seriesIndices[i] + 1,
                        type: DataProcessor.helpers.getChartType(
                            chart.seriesdata[seriesIndices[i] + 1].type,
                            'string',
                            seriesIndices[i] + 1,
                            chart.userdata
                        )
                    },
                    chart
                );
            } else {
                ele = getClassInfoForAddSeries(
                    {
                        data: chart.seriesdata[i],
                        index: i,
                        type: DataProcessor.helpers.getChartType(chart.seriesdata[i].type, 'string', i, chart.userdata)
                    },
                    chart
                );
            }
            if (!ele.oldSeriesClass) {
                ele.oldSeriesClass = [];
            }
            ele.newSeriesClass.map((d, i) => {
                createElement('g', d, [{ data: currData, index: i }], seriesholder, ele.oldSeriesClass[i], d => d);
            });
        }
    }
}

export var addSeries = async function(seriesInfo, seriesIndices) {
    var chart = this,
        nonAxisCharts = chartcategory.nonaxis;

    seriesInfo = splat(seriesInfo);
    seriesIndices = splat(seriesIndices);
    try {
        seriesInfo.some((d, i) => {
            if (nonAxisCharts.indexOf(charttype[d.type]) != -1) {
                throw `Action not possible! Add series is not supported for "${d.type}"`;
            }
        });
        chart.seriesAddRemoveDetails = updateAddRemoveSeriesProperty(
            chart.seriesAddRemoveDetails,
            'add',
            chart.seriesdata && chart.seriesdata.length
        );

        var renderingMode = (chart && chart.plot && chart.plot.renderer && chart.plot.renderer.mode) || 'svg',
            oldSeriesDataLength = chart.userdata.seriesdata.chartdata.length,
            seriesDataIndexMap = reConstructChartData(seriesInfo, seriesIndices, chart),
            flattenPoints = chart.dataObject.flattenPoints;

        //we need not insert element to DOM in canvas case and if we add series to a null data,chart.seriesdata not updated in this case
        if (renderingMode === 'svg' && oldSeriesDataLength) {
            insertSeriesToDOM(chart, seriesIndices);
        }
        chart.redraw(true);
        if (chart.seriesAddRemoveDetails.seriesUpdatedAtEnd && !chart.dataObject.flattenPoints) {
            chart.dataObject.flattenPoints = flattenPoints;
            updateFlattenPoints(seriesDataIndexMap, seriesIndices, chart);
        }

        if (chart.dataObject.pushFlattenAnnotation && chart.eventHandler.isSingleSource()) {
            chart.dataObject.pushFlattenAnnotation();
        }

        chart.eventHandler.reverseLegendHighlight = new (Registry.getComponent('legend.ReverseLegendHighlight'))(chart);
        chart.seriesAddRemoveDetails.seriesAddRemovePhase = false;
    } catch (err) {
        VizLayout.helpers.handleErrors(chart, err);
    }
};
function updateFlattenPoints(seriesDataIndexMap, seriesIndices, chart) {
    seriesIndices.map(i => {
        var currSeriesData = seriesDataIndexMap.get(i).data[0],
            currSeriesIdx = i,
            renderer = chart.renderer[currSeriesIdx],
            flattenData = {
                seriesindex: currSeriesIdx,
                subseriesindex: 0,
                row: null,
                column: null,
                index: 0,
                rendererIndex: renderer.commonRendererProp().rendererIndex()
            };
        currSeriesData.map(data => {
            flattenData.data = [[data]];
            chart.dataObject.flattenData([flattenData]);
        });
    });
}
