// @zohocharts/zc/src/core/data-processor/hierarchy/groupby-time-layout.js
import { Row } from '../../../pivot/rows';
import { Column } from '../../../pivot/columns';
import { Layer } from '../../../pivot/layers';
import { Node } from '../../../pivot/node';
import { processChild } from '../../../pivot/children';
import { setRanges } from '../../../pivot/categories-range';
import { setLeafSize } from '../../../pivot/leaf-size';
import { eachNodeBeforeAfter } from '../../../pivot/node-iterator';
import { Registry } from '../../zc-registry';

export function GroupByTimeDataLayout(data, hooks) {
    let _this = this;
    (this.rows = new Row()), (this.columns = new Column()), (this.layers = new Layer()), (this.series = []);

    var root = new Node(data, -1, -1, -1, 'root', '');

    root.children = data.data;

    eachNodeBeforeAfter(
        root,
        function(node) {
            if (node.type == 'layer') {
                node.data = [
                    node.children.map(d => {
                        return d.data;
                    })
                ];
            }
            _this.process(node);
        },
        this.postProcess
    );
    this.rows.branches.reverse();
    this.columns.branches.reverse();
}

function preProcess(node) {
    if (node.tag == 'layer') {
        node.data = [
            node.children.map(d => {
                return d.data;
            })
        ];
    }
    this.process(node);
}

function postProcess(node, child) {
    child.row = node.row || 0;
    child.column = node.column || 0;
}

function getRootNodeList(tag) {
    let list = tag == 'row' ? this.rows : this.columns;
    return list;
}

GroupByTimeDataLayout.prototype = {
    process: processChild,
    preProcess: preProcess,
    postProcess: postProcess,
    getRootNodeList: getRootNodeList,
    setRanges: setRanges,
    leafSize: setLeafSize
};

Registry.setComponent('GroupByTimeDataLayout', GroupByTimeDataLayout);
