// @zohocharts/zc/src/core/data-processor/hierarchy/hierarchy-time-data-adapter.js
// (function() {
//hierarchy-time-data-adapter.js
/**
 *
 * @param {Chart} chartInstance
 * @param {Array} tickValues
 * @param {Array} groupOrder
 * @param {Boolean} isDiscrete
 * @param {Object} columnInfo
 * @constructor
 */
import { inherit, extend, pick, isArray, defined } from '../../../utils/zcutils';
import { mathCeil, mathMin } from '../../zc-constants';
import { TIME_DATATYPE } from '../../datatype-constants';
import { getQuarter, millisecondToAbsoluteTime, generateDateVal } from '../../../formatter/formatter';
import { DataProcessor as P } from '../../data-processor';
import { scale } from '../../main-zc-init';
import { Node } from '../../../pivot/node';
import { Registry } from '../../zc-registry';

export * from './hierarchy-data-adapter';

function HierarchicalTimeDataAdapter(chartInstance, tickValues, groupOrder, isDiscrete, columnInfo) {
    P.HierarchicalDataAdapter.call(this, chartInstance);

    this.dataType = TIME_DATATYPE;
    this.chart = chartInstance;
    this.columnInfo = columnInfo;

    if (isDiscrete) {
        this.groupByDiscreteTime = true;
    } else {
        this.groupByContinuousTime = true;
    }

    if (isArray(groupOrder)) {
        this.groupOrder = groupOrder.slice(0).reverse();
    } else if (groupOrder === 'AUTO') {
        this.groupOrder = HierarchicalTimeDataAdapter.getAutoGroupOrder(columnInfo.time.subfunction);
    }

    this.groupByTime = true; // common

    this.update(tickValues);
}

var months = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'],
    days = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'],
    defaultGroupOrder = ['YEAR', 'QUARTER', 'MONTH', 'WEEK', 'DAY'],
    formattingFunctions = {
        YEAR: 'YEAR',
        ABSQUARTER: 'QUARTER',
        QUARTER: 'QUARTER',
        MONTH: 'MONTH',
        ABSMONTH: 'MONTH',
        WEEK: 'WEEK',
        ABSWEEK: 'WEEK',
        ABSDAY: 'DAY'
    };

/**
 *
 * @param {Date} d
 * @returns {string}
 */

function getMonth(d) {
    return months[d.getMonth()];
}

/**
 *
 * @param {Date} d
 * @returns {string}
 */
function getDay(d) {
    return days[d.getDay()];
}

/**
 *
 * @param {Date} d
 * @returns {number}
 */
function getWeek(d) {
    var firstJanuary = new Date(d.getFullYear(), 0, 1);
    return mathCeil(((d - firstJanuary) / 86400000 + firstJanuary.getDay() + 1) / 7);
}

/**
 *
 * @param array
 * @param callback
 * @returns {*}
 */

function groupBy(array, callback) {
    var key = null;
    return array.reduce(function(r, a) {
        key = callback.call(null, a);
        r[key] = r[key] || [];
        r[key].push(a);
        return r;
    }, {});
}

/**
 *
 * @param tickValues
 * @param columnInfo
 */
function getDateFromTickValues(tickValues, columnInfo) {
    return tickValues.map(function(d) {
        return generateDateVal(d, null, columnInfo);
    });
}

/**
 *
 * @param data
 */

function convertDateToObject(data) {
    var converter = function(point) {
        return {
            YEAR: point.getFullYear(),
            QUARTER: getQuarter(point),
            MONTH: getMonth(point),
            WEEK: getWeek(point),
            DAY: getDay(point),
            timestamps: point.getTime(),
            data: point
        };
    };
    return data.map(converter);
}

/**
 *
 * @param object
 * @param leaf
 * @param depth
 * @param isDiscreteTime
 * @returns {Array}
 */

function toHierarchy(object, leaf, depth, isDiscreteTime) {
    var tree = [];
    for (var key in object) {
        if (object.hasOwnProperty(key)) {
            var entry = new Node(object[key], depth, depth, depth);

            tree.push(entry);

            if (leaf) {
                entry.leaf = true;
                entry.value = isDiscreteTime ? object[key].data.getTime() : object[key].data;
                entry.point = object[key].data.getTime();
                entry.tag = entry.leaf ? 'leaf' : 'column';
                entry.datatype = {
                    rows: 'numeric',
                    columns: isDiscreteTime ? 'ordinal' : 'time',
                    _columns: 'time'
                };
                entry.seriesname = 'manual';
                entry.isDefined = true;
            } else {
                entry.value = key;
                entry.point = object[key][0].data;
                entry.children = object[key];
                entry.tag = entry.leaf ? 'leaf' : 'column';
                entry.datatype = 'ordinal';
                entry.depthbyrow = entry.tag == 'row' ? entry.depthbyrow + 1 : entry.depthbyrow;
                entry.depthbycolumn = entry.tag == 'column' ? depth : depth;
                entry.seriesname = 'manual';
                entry.isDefined = true;
            }
        }
    }
    return tree;
}

/**
 *
 * @type {{update: update, process: process, getPossibleGroupBy: getPossibleGroupBy}}
 */

var timeAxisGroup = {
    setScaleForRange: function(leafscale) {
        let _this = this;
        return {
            _padding: function(nodes, points) {
                let scale_copy = leafscale.copy(),
                    firstPt = nodes[0].node,
                    lastPt = nodes[nodes.length - 1].node,
                    isOrdinal = leafscale.bandwidth,
                    rangeBand = isOrdinal ? scale._getDiscreteRangeBand(leafscale) : scale.getContinuousRangeBand(leafscale, points),
                    innerPadding = isOrdinal ? rangeBand.rangeBand : rangeBand.rangeBand / 2,
                    outerPadding = rangeBand.outerPaddingInPixel;

                leafscale.firstPt = firstPt;
                leafscale.lastPt = lastPt;
                leafscale.innerPadding = innerPadding;
                leafscale.outerPadding = outerPadding;
            },
            _rangebandForFirstBranch: function(nodes) {
                var start = nodes[0],
                    end = nodes[nodes.length - 1],
                    pointProperty = 'point',
                    barWidth = leafscale.bandwidth ? leafscale.bandwidth() : null;

                if (_this.groupByContinuousTime) {
                    barWidth = 0;

                    if (defined(_this.tickValues[start.index + 1])) {
                        var startTick = _this.tickValues[start.index],
                            nextTick = _this.tickValues[start.index + 1],
                            index = start.index == 0 ? start.index : start.index - 1,
                            rangeBand = scale.getContinuousRangeBand(leafscale, _this.tickValues, index),
                            innerPadding = rangeBand.rangeBand;
                        // r = leafscale(nextTick) - leafscale(startTick);

                        leafscale.innerPadding = innerPadding;
                    }
                } else {
                    pointProperty = 'index';
                }

                return (
                    (leafscale(end[pointProperty]) -
                    leafscale(start[pointProperty]) +
                    this.getSpan(start, true) +
                    this.getSpan(end) +
                    barWidth)
                );
            },
            getSpan: function(point, startSide) {
                if ((startSide && point === leafscale.firstPt) || (!startSide && point === leafscale.lastPt)) {
                    return leafscale.outerPadding;
                } else {
                    return leafscale.innerPadding;
                }
            },
            _scale: leafscale,
            pointProperty: leafscale.bandwidth ? 'index' : 'point'
        };
    },
    update: function(tickValues) {
        var layout = this;

        if (layout.groupByDiscreteTime) {
            tickValues = pick(layout.chart.dataObject.findXCategories(), tickValues);
        }

        layout.dataCount = 0;
        layout.tickValues = getDateFromTickValues(tickValues, layout.columnInfo);
        layout.process();

        // if (!this.pivotLayout) {
        // TODO check for the need of recalculation
        var component = new Registry.getComponent('GroupByTimeDataLayout');
        this.pivotLayout = new component({ data: this.categories });
        // }
    },

    process: function() {
        var layout = this,
            data = layout.tickValues,
            index = 0,
            groupOrder = pick(layout.groupOrder, defaultGroupOrder),
            isDiscreteTime = layout.groupByDiscreteTime;

        var object = convertDateToObject(data, index),
            order = groupOrder.slice(0),
            key = order.shift(),
            yearlyGrouped = groupBy(object, function(date) {
                return date[key];
            }),
            currentDepth = 0,
            storage = toHierarchy(yearlyGrouped, false, currentDepth, isDiscreteTime);

        function group(currentObj, keys, depth) {
            var key = keys.shift(),
                grouped;
            currentObj.forEach(function(d) {
                if (key) {
                    grouped = groupBy(d.children, function(data) {
                        return data[key];
                    });

                    d.children = toHierarchy(grouped, false, depth, isDiscreteTime);
                    group(d.children, keys.slice(0), depth + 1);
                } else {
                    d.children = toHierarchy(d.children, true, depth, isDiscreteTime);
                }
            });

            return currentObj;
        }

        layout.categories = group(storage, order.slice(0), currentDepth + 1);
    },

    /**
     *
     * @param {[Date]} data array of dates
     * @param {array} groupOrder
     * @returns {string}
     */

    getPossibleGroupBy: function(data, groupOrder) {
        var startTime = data[0],
            endTime = data[data.length - 1],
            isSameYear = startTime.getFullYear() === endTime.getFullYear(),
            groupByObject = groupOrder[0];

        var timeDiff = data[data.length - 1] - data[0];
        var absoluteTime = millisecondToAbsoluteTime(timeDiff);

        for (var property in absoluteTime) {
            if (absoluteTime.hasOwnProperty(property)) {
                if (absoluteTime[property] > 1 || (property === 'YEAR' && !isSameYear)) {
                    groupByObject = property;
                    break;
                }
            }
        }
        return groupByObject;
    }
};

/**
 *
 * @param {String} formattingFunction
 * @returns {Array.<string>}
 */
HierarchicalTimeDataAdapter.getAutoGroupOrder = function(formattingFunction) {
    var index = defaultGroupOrder.indexOf(formattingFunctions[formattingFunction]);
    return defaultGroupOrder.slice(0, index);
};

// inherit from HierarchicalDataAdapter
inherit(HierarchicalTimeDataAdapter, P.HierarchicalDataAdapter);

// extend timeAxisGroup
extend(HierarchicalTimeDataAdapter.prototype, timeAxisGroup);

P.HierarchicalTimeDataAdapter = HierarchicalTimeDataAdapter;

// $ZC.HierarchicalTimeDataAdapter = HierarchicalTimeDataAdapter;

// })();

export var _hierarchyTime = P.HierarchicalTimeDataAdapter;
