// @zohocharts/zc/src/core/data-processor/hierarchy/hierarchy-ordinal-data-adapter.js
// (function() {

//hierarchy-ordinal-data-adapter.js
// "use strict";
import {
    defined,
    pick,
    inherit,
    extend,
    objectValues,
    getPropVal,
    invokeFunction,
    getDistinct, arrayIncludes
} from '../../../utils/zcutils';
import { ORDINAL_DATATYPE } from '../../datatype-constants';
import { PreProcessor } from '../pre-data-processor';
import { DataProcessor, DataProcessor as P } from '../../data-processor';
import { d3_merge, d3_map, d3_scaleOrdinal, d3_range } from '../../d3-constants';
import {chartcategory, charttype, scale} from '../../main-zc-init';
import { Registry } from '../../zc-registry';
import { getMergedDatalabelOptions } from '../../../label/datalabelsUtils';
import { getChartTypeByIndex } from '../../charttype-helpers';
import { dataLabelTextContent } from '../../../label/datalabel-content';
import { getTextBound } from '../../../axis/textBound';
import { mathFloor, mathMin, mathAbs, mathCeil } from '../../zc-constants';
import { appendEle } from '../../../utils/zcsvgutils';
import { getTextBoundByDOM } from '../../../utils/zcsvgutils';
import { ZC_getColumnSize } from '../../../axis/column-size-utils';
import {getAllSeriesTypes} from "../../series-parser-helpers";

export * from './hierarchy-data-adapter';

function HierarchicalOrdinalDataAdapter(chartInstance) {
    P.HierarchicalDataAdapter.call(this, chartInstance);
    this.dataType = ORDINAL_DATATYPE;
    this.categoryTree = true;
    this.gridOptions = {};
    this.legendOptions = {};
    this.update(true);
}

/**
 *
 * @type {{process: process, getSeries: getSeries, searchHierarchy: searchHierarchy}}
 */

var hierarchicalOrdinalDataAdapter = {
    calculatePivotLayout: function() {
        var helpers = this.pivotHelpers(),
            component = this.pivot ? Registry.getComponent('PivotLayout') : new Registry.getComponent('LayerHierarchyLayout'),
            pivot = this.pivot
                ? new component({ seriesname: 'root', data: this.originalSeries }, helpers)
                : new component(this.originalSeries, helpers);

        this.pivotLayout = pivot;
    },
    hasRows: function() {
        let rowList = this.pivotLayout.getRootNodeList('row'),
            rowCount = rowList.getCount(),
            hasMultiCluster = rowList.hasMultiCluster();

        return rowCount || hasMultiCluster;
    },
    /*
    @param scale
    hooks for setting ranges for column with leafscale
     */
    setScaleForRange: function(leafscale) {
        var _this = this,
            chart = this.chart;
        return {
            _padding: function(nodes, points, index) {
                let scale_copy = leafscale.copy(),
                    firstPt = nodes[0].node,
                    lastPt = nodes[nodes.length - 1].node,
                    isOrdinal = leafscale.bandwidth,
                    rangeBand = isOrdinal ? scale._getDiscreteRangeBand(leafscale, index) : scale.getContinuousRangeBand(leafscale, points),
                    innerPadding = isOrdinal ? rangeBand.rangeBand : rangeBand.rangeBand / 2,
                    outerPadding = rangeBand.outerPaddingInPixel;

                leafscale.firstPt = firstPt;
                leafscale.lastPt = lastPt;
                leafscale.innerPadding = innerPadding;
                leafscale.outerPadding = outerPadding;
                leafscale.span = isOrdinal ? leafscale.bandwidth() : null;

                if (!defined(index)) {
                    leafscale._innerPadding = innerPadding;
                    leafscale._outerPadding = outerPadding;
                }
            },

            _scale: leafscale,
            pointProperty: leafscale.bandwidth ? 'index' : 'point',
            _rangeInfo: function() {
                var isHorizontal = chart.axes.x.orient == 'top' || chart.axes.x.orient == 'bottom',
                    startPos = isHorizontal ? 0 : chart.plotarea.height,
                    sign = isHorizontal ? 1 : -1;
                return {
                    sign: sign,
                    startPos: startPos
                };
            },
            _hasDynamicLeafScale: function() {
                return _this.dynamicLeafScale;
            },
            _haUniformLeafScale: chart.systemConf.chart.axes.xaxis.uniformLeafScale
        };
    },
    setGridOptions: function() {
        var hierarchical = this,
            chartObj = hierarchical.chart,
            gridOptions = hierarchical.gridOptions,
            legendOptions = hierarchical.legendOptions,
            pivotLayout = hierarchical.pivotLayout,
            layers = pivotLayout.layers,
            hasMultiLayers = layers.isLayerNode,
            rowList = hierarchical.pivotLayout.getRootNodeList('row'),
            hasMultiCluster = rowList.hasMultiCluster(),
            panes = layers.panes,
            gridlist = objectValues(panes),
            globalChartType = hierarchical.chart.seriesdata && hierarchical.chart.userdata.seriesdata.type,
            nonAxisSingleSeries = !(chartObj.dataObject.isAxisCategory || chartObj.dataObject.isNonAxisMultiSeriesCategory);
        gridlist.forEach((grid, gridIdx) => {
            var colors = new Array(0),
                chartTypes = new Array(0),
                legendIndices = new Array(0);
            var paneid = `${grid.row}_${grid.column}`;
            var legendNewConfig = chartObj.systemConf.legend.color && chartObj.systemConf.legend.color[grid.layer],
                legendColors = (legendNewConfig && legendNewConfig.colors) || [];
            var layerIdx = grid.layer,
                layers = hierarchical.pivotLayout.layers.groups[layerIdx];
            var data = grid.data;
            data.forEach((d, i) => {
                var legendIdx = layers.indexOf(d.value);
                var idx = hasMultiCluster ? ((hasMultiLayers ? legendIdx : gridIdx)) : legendIdx;
                colors.push(
                    d.node.color || legendColors[i] || hierarchical.chart.mainSeriesColor[idx % hierarchical.chart.mainSeriesColor.length]
                );
                var defaultType = hierarchical.chart.chartTypes[i] ? hierarchical.chart.chartTypes[i].value : globalChartType;
                chartTypes.push(P.helpers.getChartType(pick(d.node.type, globalChartType), 'object'));
                legendIndices.push(idx);
            });
            gridOptions[paneid] = gridOptions[paneid] || {};
            gridOptions[paneid].renderer = new Array(0);
            gridOptions[paneid].colors = colors;
            gridOptions[paneid].chartTypes = chartTypes;
            gridOptions[paneid].legendOrder = grid.layer;
            gridOptions[paneid].legendIndices = legendIndices;
            if (!legendOptions[grid.layer]) {
                legendOptions[grid.layer] = {
                    colors: colors
                };
            }
        });
    },
    setDummyDomainForClusters: function(columnSizeInfo, leafScale) {
        var hierarchical = this,
            chartObj = hierarchical.chart,
            pivotLayout = hierarchical.pivotLayout,
            nodeList = pivotLayout.getRootNodeList('column'),
            defaultSize = 150,
            totalCustomLeafSize = 0,
            customCount = 0;

        //Find total leaf size generated
        columnSizeInfo.map(d => {
            if (d && d !== 'auto') {
                totalCustomLeafSize += d;
                customCount++;
            }
        });
        //Find available plot size
        var scaleRange = leafScale.range();
        var plotSize = mathAbs(scaleRange[1] - scaleRange[0]),
            newPlotSize = plotSize - totalCustomLeafSize;

        var tempScale = d3_scaleOrdinal()
            .domain(d3_range(leafScale.domain().length - customCount))
            .range([0, newPlotSize]);
        // var newCSize = (newPlotSize / (leafScale.domain().length - customCount));
        var bw = tempScale.bandwidth();
        columnSizeInfo.map((d, i) => {
            if (d && d !== 'auto') {
                var childCount = nodeList.getNodesForClusterSet(0, i).length;
                var newCSize = bw * childCount;
                hierarchical.dynamicLeafScale = true;
                var extendedCategoriesCount = 0,
                    maxColumnSize = d;
                if (maxColumnSize > newCSize) {
                    extendedCategoriesCount = mathMin(mathCeil(maxColumnSize / newCSize), 5);
                }

                //add dummy category in pivot entries
                if (extendedCategoriesCount) {
                    var clusterSet = nodeList.clusters[0].entries.get(`${i}_`);
                    clusterSet.customDomain = d3_map();
                    var customDomain = new Array(extendedCategoriesCount);
                    var mapObj = {
                        count: extendedCategoriesCount,
                        values: customDomain
                    };
                    for (var i = 0; i < extendedCategoriesCount; i++) {
                        customDomain[i] = `domain_${i}`;
                    }
                    clusterSet.customDomain.set(0, mapObj);
                }
            }
        });
    },
    getDynamicColumnSize: function(leafScale, customSize) {
        var hierarchical = this,
            chartObj = hierarchical.chart,
            pivotLayout = hierarchical.pivotLayout,
            nodeList = pivotLayout.getRootNodeList('column'),
            leafSizeCallback = chartObj.systemConf.facets.config.columns.leafSize,
            columnInfo;
        //TOdo
        customSize = chartObj.plotarea.width;

        if (nodeList.count || nodeList.leafCount) {
            //start
            let datatype = nodeList.clusters[0].datatype;
            if (datatype == ORDINAL_DATATYPE) {
                if (nodeList.count) {
                    var nodeData = nodeList.branches[0].domain.map((d, i) => {
                        return nodeList.getNodesForClusterSet(0, i);
                    });
                    var info = {
                        nodes: nodeData,
                        count: nodeData.length
                    };
                    if (leafSizeCallback == 'default') {
                        columnInfo = ZC_getColumnSize(info, leafScale, chartObj);
                    } else {
                        columnInfo = invokeFunction(leafSizeCallback, info, leafScale, chartObj);
                    }
                    this.setDummyDomainForClusters(columnInfo, leafScale);
                }
                //end
                return;
            }
        }
    },
    getNodeValueByColumnIdx: function(point, colidx, levelindex) {
        let _this = this,
            chart = this.chart,
            facets = chart.userdata.facets,
            rowIndices = chart.dataset.rowIndices,
            columnIndices = chart.dataset.columnIndices;
        var tag, depth, inverseDepth, nodeIdx;
        if (chart.dataObject.pivot) {
            var rowDepth = rowIndices.indexOf(colidx);
            if (rowDepth != -1) {
                tag = 'row';
                depth = rowDepth;
                inverseDepth = rowIndices.length - 1 - rowDepth;
                nodeIdx = point.levelByRow.split('_')[depth];
            } else {
                var columnDepth = columnIndices.indexOf(colidx);
                tag = 'column';
                depth = columnDepth;
                inverseDepth = columnIndices.length - 1 - columnDepth;
                nodeIdx = point.levelByColumn.split('_')[depth];
            }
        } else {
            tag = 'column';
            depth = levelindex - 1;
            inverseDepth = chart.axes.x.count - 1 - levelindex;
            nodeIdx = point.levelByColumn.split('_')[depth];
        }
        var nodeList = chart.dataObject.hierarchical.pivotLayout.getRootNodeList(tag),
            domainList = nodeList.getDomainForBranch(inverseDepth),
            nodeValue = domainList[nodeIdx];

        return nodeValue;
    },
    /*
    Hooks used for both pivot and layer hierarchy model
     */
    pivotHelpers: function() {
        let _this = this,
            chart = this.chart,
            allDistinctSeriesTypes_val = chart.distinctChartTypes || getDistinct(getAllSeriesTypes(chart.userdata)),
            allChartTypes_num = allDistinctSeriesTypes_val.map(_ => charttype[_]);
        var helpers = {};

        helpers.children = node => {
            var hasChildren = PreProcessor.helpers.hasInnerSeries(node);
            if (hasChildren) {
                return node.data;
            } else if (node.data) {
                var is3d = P.helpers.is3D(node.data),
                    is1d = !is3d && node.data.length && node.data[0].constructor !== Array,
                    nodeData = is3d ? d3_merge(node.data) : is1d ? [node.data] : node.data,
                    hasSomeDefinedData = nodeData.length
                        ? nodeData.some(d => {
                              return helpers.isDefined(d);
                          })
                        : true,
                    isEmpty = !nodeData.length || (nodeData.length == 1 && !nodeData[0].length) || !hasSomeDefinedData;

                return isEmpty ? null : nodeData;
            }
            return null;
        };
        (helpers.nodeType = (node, depth1) => {
            var depth = node ? pick(depth1, node.depth) : depth1,
                axes = chart.userdata.metadata.axes,
                yindices = d3_merge(axes.y),
                columns = chart.userdata.metadata.columns,
                facets = chart.userdata.facets,
                mappedIndices = mapColumnIndexToLevelIndex(facets, columns),
                rowIndices = facets
                    ? chart.dataset.rowIndices ||
                      facets.rows.filter(d => {
                          return yindices.indexOf(d) == -1;
                      })
                    : mappedIndices.rows,
                columnIndices = facets
                    ? chart.dataset.columnIndices ||
                      facets.columns.filter(d => {
                          return axes.x.indexOf(d) == -1;
                      })
                    : mappedIndices.columns,
                columnIdx = mappedIndices.columns.indexOf(depth),
                rowIdx = mappedIndices.rows.indexOf(depth),
                layerIdx = mappedIndices.color.indexOf(depth),
                isColumn = columnIdx != -1,
                isRow = rowIdx != -1,
                isLayer = layerIdx != -1; //TODO alt for indexof

            chart.dataset.rowIndices = rowIndices;
            chart.dataset.columnIndices = columnIndices;

            var hasChildren = node ? node.node.data : true; //PreProcessor.helpers.hasInnerSeries(tree);

            if (node) {
                /*
                        In pivot model, if row/column datatype is ordinal and has no child, we need to process
                        separately by pushing to leaves.
                     */
                var hasGrandNode = PreProcessor.helpers.hasInnerSeries(node);
                node.hasInnerChild = true;
                if (_this.pivot) {
                    node.hasInnerChild = hasGrandNode;
                }
            }

            if (isColumn && hasChildren) {
                var isFirstColumn = columnIdx == 0,
                    isLastColumn = columnIdx == columnIndices.length - 1;
                if (node && isFirstColumn) {
                    node.firstColumn = isFirstColumn;
                }
                if (node && isLastColumn) {
                    node.lastColumn = isLastColumn;
                }

                return 'column';
            } else if (isRow && hasChildren) {
                var isFirstRow = rowIdx == 0,
                    isLastRow = rowIdx == rowIndices.length - 1;
                if (node && isFirstRow) {
                    node.firstRow = isFirstRow;
                }
                if (node && isLastRow) {
                    node.lastRow = isLastRow;
                }

                if (node && !node.hasInnerChild && mappedIndices.color.length) {
                    node.mainTag = 'layer';
                }
                return 'row';
            } else if (isLayer && hasChildren) {
                if (node) {
                    node.layer = facets && facets.color;
                }
                return 'layer';
            } else {
                return 'leaf';
            }
        }),
            (helpers.data = function(node) {
                var hasChildren = PreProcessor.helpers.hasInnerSeries(node);
                if (hasChildren) {
                    return node.data;
                } else if (node.data) {
                    var is3d = P.helpers.is3D(node.data),
                        is1d = !is3d && node.data.length && node.data[0].constructor !== Array,
                        nodeData = is3d ? node.data : is1d ? [[node.data]] : [node.data];
                    node.data = nodeData;
                    return nodeData;
                } else {
                    return node;
                }
            }),
            (helpers.isEmpty = function(node) {
                var hasChildren = helpers.children(node);
                return !hasChildren;
            }),
            (helpers.dataType = (node, depth1) => {
                var type = node.tag,
                    depth = pick(depth1, node.depth),
                    axes = chart.userdata.metadata.axes,
                    columns = chart.userdata.metadata.columns,
                    facets = chart.userdata.facets,
                    mappedIndices = mapLevelIndexToColumnIndex(facets, columns, axes),
                    columnindex = mappedIndices.get(depth),
                    datatype = columns[columnindex || 0].datatype,
                    isBarType = allChartTypes_num.every(chartType => arrayIncludes(chartcategory.discrete, chartType));
                if (type == 'leaf') {
                    let columnInfo_rows = columns[axes.y[0][0]],
                        columnInfo_columns = columns[axes.x[0]];
                    datatype = {
                        rows: columnInfo_rows.useDiscrete ? ORDINAL_DATATYPE : columnInfo_rows.datatype,
                        columns: columnInfo_columns.useDiscrete || isBarType ? ORDINAL_DATATYPE : columnInfo_columns.datatype
                    };
                }

                return datatype;
            }),
            (helpers.nodeName = function(node) {
                return node.seriesname || node[0];
                // if(node.seriesname){
                //     return node.seriesname;
                // }else{
                //     let column = pick(chart.dataset.getColumn('x'), {}),
                //         dataType = column.datatype,
                //         dataIndex = pick(column.dataindex, 0),
                //         hasNullFormatter = defined(column[dataType]) && defined(column[dataType].nullformatting),
                //         nullFormatter = hasNullFormatter ? column[dataType].nullformatting : null,
                //         value = node[dataIndex] || nullFormatter;
                //     return value;
                // }
            }),
            (helpers.layerName = function(node) {
                return node.seriesname;
            }),
            (helpers.isDisabledSeries = function(series) {
                return series.disabled;
            }),
            (helpers.isDefined = function(data) {
                let column = pick(chart.dataset.getColumn('x'), {}),
                    dataType = column.datatype,
                    dataIndex = pick(column.dataindex, 0),
                    hasNullFormatter = defined(column[dataType]) && defined(column[dataType].nullformatting);
                return hasNullFormatter || defined(data[dataIndex]);
            }),
            (helpers.hasSplitLeavesColumns = function() {
                if (chart.systemConf) {
                    return chart.systemConf.chart.axes.xaxis.splitLeaves;
                }
                return getPropVal(chart.userdata, 'chart.axes.xaxis.splitLeaves');
            });

        return helpers;
    },
    update: function(init) {
        if (!init) {
            this.initialize();
        }
        this.pivot = this.isPivotModel(this.originalSeries);
        this.facetRendering = this.renderByFacet();
        this.calculatePivotLayout();
    },
    /*
    To check whether data model is pivot or of layer hierarchy
     */
    isPivotModel: function(root) {
        return this.pivotHelpers().nodeType(null, 0) != 'layer';
    },
    /*
    To check whether plot need to render by facet columns
     */
    renderByFacet: function() {
        return (this.pivot) || (this.chart.userdata.facets && this.chart.userdata.facets.renderByFacet);
    },
    getSeries: function(seriesIndex) {
        return defined(seriesIndex) ? this.pivotLayout.series[seriesIndex] : this.pivotLayout.series;
    }
};

/**
 *
 * @param {facets, columns}
 * @Generate map column indices to hierarchical depth of the data
 */
function mapColumnIndexToLevelIndex(facets, columns) {
    var mappedIndices = {
            rows: [],
            columns: [],
            color: []
        },
        arr = ['rows', 'columns', 'color'];
    if (facets) {
        arr.map(d => {
            if (facets[d]) {
                facets[d].map(i => {
                    mappedIndices[d].push(columns[i].levelindex);
                });
            }
        });
    } else {
        mappedIndices.color.push(0); //default 0 for color indices;
        columns.map(d => {
            if (d.levelindex) {
                mappedIndices.columns.push(d.levelindex);
            }
        });
    }

    return mappedIndices;
}
/**
 *
 * @param {facets, columns}
 * @Generate map hierarchical depth to column indices
 */
function mapLevelIndexToColumnIndex(facets, columns, axes) {
    var mappedIndices = d3_map(),
        arr = ['rows', 'columns', 'color'];
    if (facets) {
        arr.map(d => {
            if (facets[d]) {
                facets[d].map(i => {
                    mappedIndices.set(columns[i].levelindex, i);
                });
            }
        });
    } else {
        mappedIndices.set(0, axes.clr[0]);
        columns.map((d, i) => {
            if (d.levelindex) {
                mappedIndices.set(d.levelindex, i);
            }
        });
    }

    return mappedIndices;
}

// inherit from HierarchicalDataAdapter
inherit(HierarchicalOrdinalDataAdapter, P.HierarchicalDataAdapter);

// extend all the functions
extend(HierarchicalOrdinalDataAdapter.prototype, hierarchicalOrdinalDataAdapter);
P.HierarchicalOrdinalDataAdapter = HierarchicalOrdinalDataAdapter;

Registry.setComponent('_hierarchyOrdinal', P.HierarchicalOrdinalDataAdapter);
export var _hierarchyOrdinal = P.HierarchicalOrdinalDataAdapter;
