// @zohocharts/zc/src/events/visualEffects/behaviour/behaviour-map.js
import { Registry } from "../../../core/zc-registry";

class BehaviourMap {
    constructor() {
        //TODO: Change to private member later
        // if (this.instance) {
        //     return this.instance;
        // }
        this.__behaviour = new Map();
    }
    list() {
        return this.__behaviour.keys().toArray();
    }

    /**
     * 
     * @param {behaviour} behaviour
     */
    set(behaviour = {}) {
        const { type, action, ...options } = behaviour;
        console.assert(typeof type === 'string', 'TypeError: behaviour.type must be a string');
        console.assert(typeof action === 'function', 'TypeError: behaviour.action must be a function');

        this.__behaviour.set(behaviour.type, new Behaviour(type, action, options));
    }
    /**
     * 
     * @param {string} key - behaviour.type
     */
    get(key) {
        let behaviour = this.__behaviour.get(key);
        console.assert(behaviour, `TypeError: '${key}' not found in Behaviour map. Refer BehaviourMap.list to get the list of behaviours available.`);
        const { type, action, ...options } = behaviour;
        return new Behaviour(type, action, options);
    }
}

/**
 * 
 * @param {behaviour['type']} type
 * @param {behaviour['action']} action
 * @param {behaviour} options
 */
class Behaviour {
    constructor(type, action, options = {}) {
        for (const key in options) {
            this[key] = options[key]
        }

        this.type = type;
        this.action = action;
    }
}

const behaviourMap = new BehaviourMap();
Registry.setComponent('BehaviourMap', behaviourMap);
export { behaviourMap as BehaviourMap }

/**
 * @typedef {Object} behaviour
 * @property {String} behaviour.type - name of the behaviour
 * @property {Function} behaviour.action - action to be performed by the behaviour. The function must return a callback function
 * @property {Boolean} behaviour.allowDefault - action to be performed when default is true
 * @property {Boolean | Function} behaviour.allowifcustom - action to be performed when custom value is true
 */
