// @zohocharts/zc/src/bullet/bullet-levelmarker-encode.js
import { plot } from '../core/main-zc-init';
import { mathMin } from '../core/zc-constants';
import { isMultiColoring } from '../utils/colorutils';
import { isArray, pick } from '../utils/zcutils';
import { getAbsRectCoords, getAbsRectPartialCoords } from '../bar/bar-encode';
import { _setOpacityForTickFilterElements } from '../plot/svg/set-pivot-attr';
import { d3_select } from '../core/d3-constants';

export let bullet_levelmarker_position = function() {
    const KEY = 'levelMarker';
    var _this = this;

    let borderComponent = this.builderC.border;

    let position_helpers = {};

    const axesRotated = this.chart.axes.rotated,
        yaxisreversed = this.builderC.yaxisreversed;

    let builderC = this.builderC,
        cType = this.type;

    if (this.builderC[KEY] && this.builderC[KEY].enabled) {
        this.helpers.position[KEY] = {
            start: (d, encode) => {
                let { stacked, hasNegValue, yaxiscolumnorder } = this.builderC;
                let chartPhase = this.chart.phase;
                let x, y, width, height;
                let radius = borderComponent && borderComponent.show ? borderComponent.radius : null;
        
                let x_data = this.chart.dataset.getX(d),
                    yRange_low = this.scales.y.range()[0],
                    y_negValue = hasNegValue != null ? this.scales.y(hasNegValue) : null,
                    rangeY_0 = this.rangeYType ? this.scales.y(this.chart.dataset.getY(d, 0, 0)) : null,
                    // rangeY_0 = this.rangeYType ? this.scales.y(this.chart.dataset.getY(d, yaxiscolumnorder, 0)) : null,
                    y_priority = pick(rangeY_0, y_negValue);
        
                x = stacked
                    ? this.scales.x(x_data) + this.scales.interX(this.commonOption.stackGroupIndex) + this.commonOption.padding
                    : this.scales.x(x_data) + this.scales.interX(this.barDatasetIndex + (d.subseriesIdx || 0)) + this.commonOption.padding;
                if (this.scales.y.bandwidth) {
                    y = this.scales.y.origRange[0];
                    // y = this.rangeYType ? rangeY_0 + this.scales.y.bandwidth() / 2 : this.scales.y.origRange[0];
                } else {
                    if (y_priority != null) {
                        y = y_priority;
                    } else {
                        if (stacked && chartPhase == 'legendFilter') {
                            yRange_low = this.scales.y(d.y0);
                        }
                        let y_zeroValue = this.scales.y(0);
                        y = this.chart.axes.rotated || stacked ? yRange_low : y_zeroValue != null ? mathMin(y_zeroValue, yRange_low) : yRange_low;
                    }
                }
        
                width = this.chart.axes.rotated ? 0 : this.commonOption.barWidth;
                height = this.chart.axes.rotated ? this.commonOption.barWidth : 0;
        
                ({x, y, width, height} = this.helpers.positionComponent[KEY].coordinates(d, this.builderC.datasetIndex, {x, y, width, height}));
                let rect = this.chart.axes.rotated ? { x: y, y: x, width, height, radius } : { x, y, width, height, radius };
        
                if (isNaN(x)) {
                    return getAbsRectPartialCoords(rect, this.builderC.yaxisreversed, this.chart.axes.rotated, this.chart.axes.rotated ? 'y' : 'x');
                }
                if (isNaN(y)) {
                    return getAbsRectPartialCoords(rect, this.builderC.yaxisreversed, this.chart.axes.rotated, this.chart.axes.rotated ? 'x' : 'y');
                }
                if (!encode) {
                    let coords = getAbsRectCoords(rect, this.builderC.yaxisreversed, this.chart.axes.rotated);
                    return coords;
                }
            },
            end: (d, encode, relative) => {
                // let radius = null;
                // radius = this.helpers.positionComponent[KEY].radius(d, this.builderC.datasetIndex, radius);
                // if(encode == 'radius') {
                //     return radius;
                // }
                let bullet_coords = plot.render.bulletHelpers.ZC_BAR_getLevelMarkerRect(
                        d,
                        this.builderC.levelMarker,
                        this.chart,
                        this.commonOption,
                        this.builderC,
                        d.subseriesIdx || 0,
                        this.builderC.datasetIndex
                    ),
                    rect_coords;
                if (this.chart.axes.rotated) {
                    rect_coords = {
                        x: bullet_coords.y,
                        y: bullet_coords.x,
                        width: bullet_coords.width,
                        height: bullet_coords.height
                    };
                } else {
                    rect_coords = bullet_coords;
                }
                if (relative) {
                    return rect_coords;
                }
                let radius = this.helpers.position.bar.end(null, 'radius');
                radius = this.helpers.positionComponent[KEY].radius(d, this.builderC.datasetIndex, radius);
                rect_coords.radius = radius;

                rect_coords = this.helpers.positionComponent[KEY].coordinates(d, this.builderC.datasetIndex, {...rect_coords});
                let coords_abs = getAbsRectCoords(rect_coords, this.builderC.yaxisreversed, this.chart.axes.rotated);
                return coords_abs;
            },

            stop_1: (d, encode, yRangePos) => {
                let start = this.helpers.position[KEY].start(d),
                    end = this.helpers.position[KEY].end(d);
                let x1, y1, x2, y2;
                if (this.chart.axes.rotated) {
                    y1 = end.y1;
                    y2 = end.y2;
                    return { y1, y2 };
                } else {
                    x1 = end.x1;
                    x2 = end.x2;
                    return { x1, x2 };
                }
                // let coords = getAbsRectCoords({ x, y, width, height }, this.builderC.yaxisreversed, this.chart.axes.rotated);
                return { x1, y1, x2, y2 };
            },
            exit: (d, encode) => {
                const coords = { ...d.oldCoordinates };
                axesRotated ? (yaxisreversed ? (coords.x1 = coords.x2) : (coords.x2 = coords.x1))
                    : (yaxisreversed ? (coords.y2 = coords.y1) : (coords.y1 = coords.y2));
        
                return coords;
            }
            
        };
    }

    return position_helpers;
};

export let bullet_levelmarker_styles = function() {
    const KEY = 'levelMarker';
    var _this = this;
    let chart = this.chart,
        options = this.builderC[KEY],
        multiColoring = isMultiColoring(this.config.common, this.chart),
        patternPallete = this.chart.patternPallete && this.chart.patternPallete.length;

    this.helpers.positionComponent[KEY] = {};
    this.helpers.positionComponent[KEY] = {
        radius: (d, i, defaultRadius) => {
            return defaultRadius
        },
        coordinates: (d, i, defaultCoordinates = {x, y, width, height}) => {
            return defaultCoordinates
        }
    }

    if (this.builderC[KEY] && this.builderC[KEY].enabled) {
        let cType = _this.type
        options.overlayRange = 'below';
        this.helpers.styles[KEY] = {
            fill: (d, i) => {
                let itemPos = this.chart.systemConf.chart.plot.allowDuplicateCategories
                    ? this.builderC.xcategoriesReversed.indexOf(this.chart.dataset.getX(d))
                    : pick(d.sortedItemIndex, i);
                d.categoryIndex = itemPos;
                let baseClrIdx = multiColoring ? itemPos : this.legendOrder;
                var baseClr = multiColoring ? this.chart.seriesColor[baseClrIdx % this.chart.seriesColor.length] : this.builderC.color;
                let dim = patternPallete ? this.helpers.position.levelMarker.end(d, true) : null;
                var clr = isArray(options.color) ? options.color[0] : options.color;
                if (patternPallete) {
                    return this.helpers.styleComponent.bar.fillComponent(baseClrIdx, null, null, clr, null, null, dim);
                }
                return (
                    plot.render.bulletHelpers.getBulletLevelMarkerGradient(d, options, this.chart, baseClr, options.color) || clr || baseClr
                );
            },
            fillOpacity: options.fillOpacity,
            opacity: function(d, i) {
                let opacity = _setOpacityForTickFilterElements(d, i, chart, 'bar');
                d3_select(this).attr('filtered-opacity', opacity);
                return opacity;
            }
        };
    }
};
