// @zohocharts/zc/src/bullet/bullet-targetmarker-encode.js
import { getBarMarkerSize } from '../bar/bar-marker-helpers';
import { pick, defined } from '../utils/zcutils';

export let bullet_targetmarker_position = function() {
    const KEY = 'targetMarker';

    let position_helpers = {};

    let builderC = this.builderC,
        targetMarker = builderC[KEY],
        targetMarkerIdx = targetMarker.dataindex,
        barMSize_outer = getBarMarkerSize.call(this, KEY, 'outer'),
        barMSize_inner = getBarMarkerSize.call(this, KEY, 'inner');

    if (this.builderC[KEY] && this.builderC[KEY].enabled) {
        this.helpers.position[KEY] = {
            start: (d, encode) => {},
            end: (d, encode, order, prefix) => {
                if (encode == 'size') {
                    return prefix == 'outer' ? [barMSize_outer, null, null] : [barMSize_inner, null, null];
                }
                let x = this.scales.x(this.chart.dataset.getX(d)) + this.scales.x.bandwidth() / 2;

                let y = this.scales.y(d[targetMarkerIdx]);

                return {
                    x,
                    y,
                    rotate: this.chart.rotated ? 90 : 0
                };
            }
        };
    }

    return position_helpers;
};

export function bullet_targetmarker_styles() {
    const KEY = 'targetMarker';

    let options = this.builderC[KEY],
        targetMarkerIdx = options.dataindex;
    if (this.builderC[KEY] && this.builderC[KEY].enabled) {
        this.helpers.styles[KEY] = {
            fill: options.color,
            stroke: options.color,
            strokeWidth: pick(options.strokeWidth, 1),
            strokeDasharray: options.dashStyle,
            strokeLinecap: options.lineCap,
            display: (d, i) => {
                return defined(d[targetMarkerIdx]) && d[targetMarkerIdx] != null ? '' : 'none';
            }
        };
    }
}
