// @zohocharts/zc/src/bullet/bulletOptions.js
import { defaultTheme as options } from '../core/defaulttheme';

options.bullet = function(userProp) {
    return {
        strictBandWidth: true,
        outerPadding: 0.00000001,
        levelMarker: {
            enabled: true,
            dataindex: 2,
            color: [],
            colorGamma: [0.3, 0.9],
            fillOpacity: 0.9,
            maxBandWidth: 200
        },
        targetMarker: {
            enabled: true,
            dataindex: 3,
            symbol: 'line',
            innerFillColor: 'transparent',
            innerStrokeColor: 'transparent',
            innerStrokeWidth: 0,
            innerStrokeOpacity: 0,
            outerFillColor: 'rgba(0, 0, 0, 0.6)',
            outerStrokeOpacity: 1,
            outerStrokeWidth: 5,
            outerStrokeColor: 'rgba(0, 0, 0, 0.6)',
            label: {
                fontSize: userProp.datalabels.fontSize.l5
            }
        },
        animation: {
            type: 'verticalAll'
        },
        datalabels: {
            textOverlap: 'none',
            showAs: 'y', //x,y,percent
            stackLabels: {
                show: false,
                showAs: 'y' //y,percent
            }
        },
        gradients: {
            // linear || none
            options: {
                linear: {
                    y2: 50
                }
            }
        },
        stacked: {
            enabled: true
        }
    };
};

export var _bulletOptions = options.bullet;
