// @zohocharts/zc/src/bullet/bullet-levelmarker-marks.js
import { ParserRegistry } from '../core/parser-registry';
import { getAnimationConf } from '../utils/animationutils';

export function _levelMarkerMark(data, phase, condition) {
    let KEY = 'levelMarker';
    let _this = this,
        animationEnabled = _this.builderC.levelMarker?.animation?.enabled,
        builderC = this.builderC,
        attrs;

    var childOrder = _this.type =='gantt' ? null : ':first-child';

    if (this.chart.dataObject.hierarchical) {
        attrs = {
            levelbycolumn: d => d.levelByColumn,
            levelbyrow: d => d.levelByRow
        };
    }

    let animationOptions = {}
        animationOptions.animation = builderC.animationOptions;

    let plotoptions = this?.chart?.userdata?.chart?.plot?.plotoptions,
        levelMarker = plotoptions ? plotoptions[_this.type]?.levelMarker : {},
        anime_config = levelMarker ? getAnimationConf(animationOptions, levelMarker, this.chart, _this.type) : {};

    anime_config._duration = anime_config?.enabled ? anime_config.duration : 0;
    anime_config._delay = anime_config?.delay ? anime_config.delay : 0;
    anime_config._exitduration = anime_config.enabled ? 100 : 0;

    let enterMarks, updateMarks, exitMarks;

    enterMarks = {
        d: d => _this.helpers.position[KEY].start(d)
    };
    updateMarks = [
        {
            duration: _this.phase == 'redraw' ? _this.chart.phase == 'legendFilter' ?
                anime_config._duration - anime_config._exitduration : anime_config._duration : 0,
            delay: _this.chart.phase == 'legendFilter' ? anime_config._exitduration : anime_config._delay,
            d: d => _this.helpers.position[KEY].stop_1(d)
        },
        {
            duration: anime_config._duration ?? 0,
            delay: anime_config._delay,
            easing: anime_config.easingType,
            d: (d, i) => _this.helpers.position[KEY].end(d)
        }
    ];
    exitMarks = {
        duration: anime_config._exitduration,
        d: d => _this.helpers.position[KEY].exit(d)
    };


    let _levelMarkerG = {
        mark: 'g',
        order: childOrder,
        selector: '.subseries_levelMarker',
        properties: {
            attrs: {
                class: (d, i) => {
                    d.subseriesIdx = i;
                    return `subseries_levelMarker subseries_levelMarker_${i}`
                }
            }
        },
        data: function(d) {
            return condition ? (_this.builderC.stacked ? data.data : !data.disabled ? data.data : []) : [];
        },
        children: []
    };
    if (condition) {
        _levelMarkerG.children.push({
            mark: '_rect',
            selector: '.zc_levelMarker',
            data: function(d, i) {
                return _this.helpers.filterData(d);
            },
            dataAccessor: _ => _this.chart.dataset.getX(_),
            config: {
                enter: enterMarks,
                // rect: d => _this.helpers.position.levelMarker.end(),
                // d: d => _this.helpers.position.levelMarker.end(d)
            },
            anime: {
                update: updateMarks,
                exit: exitMarks
            },
            properties: {
                attrs,
                styles: {
                    fill: (d, i) => _this.helpers.styles.levelMarker.fill(d, i),
                    'fill-opacity': _this.helpers.styles.levelMarker.fillOpacity,
                    opacity: _this.helpers.styles.levelMarker.opacity
                }
            }
        });
    }
    return _levelMarkerG;
}

ParserRegistry.marks('bullet.levelmarker', _levelMarkerMark);
