// @zohocharts/zc/src/scatter/parser/scatter-parser-helpers.js
import { d3_values, defined } from '../../utils/zcutils';
import { ORDINAL_DATATYPE } from '../../core/datatype-constants';
import { DataProcessor } from '../../core/data-processor';
import { applyClippathToLivePlot } from '../../plot/dynamic-live-update/dynamic-plot-clippath';

export let scatter_parser_helpers = function () {
    let chart = this.chart,
        processor = chart.dataObject,
        dataset = chart.dataset;
    let helpers = {
        column: (type, order) => {
            let columnIdx = dataset.getColumnIdx(type, order);
            return chart.systemConf.metadata.columns[columnIdx];
        },
        datatype: (type, order) => dataset.getDataType(type, order),
        filterData: data => {
            let subseriesIdx = data.subseriesIdx || 0,
                liveUpdate = this.builderC.liveUpdate; //TODO if flatten data calculated first, we can remove this property
            if (this.seriesdataAfterUpdate[subseriesIdx] && !liveUpdate) {
                return this.seriesdataAfterUpdate[subseriesIdx];
            }

            if (liveUpdate) {
                applyClippathToLivePlot(data, this.placeHolder, this, chart);
                return DataProcessor.helpers.getDynamicData(data, this.chart);
            }

            //updateseriesdataby categories
            //renderVisiblePointsOnly
            //xAxisReverseed

            const xscale = this.scales.x,
                yscale = this.scales.y,
                yaxiscolumnorder = this.commonOption.yaxiscolumnorder,
                dataProcessorHelpers = DataProcessor.helpers;;

            let datatypeX = chart.dataset.getDataType('x'),
                datatypeY = chart.dataset.getDataType('y', yaxiscolumnorder),
                xCategories = datatypeX == ORDINAL_DATATYPE ? xscale.domain() : null,
                yCategories = datatypeY == ORDINAL_DATATYPE ? yscale.domain() : null,
                categories = datatypeX == ORDINAL_DATATYPE && xscale.domain != null && !defined(processor.hierarchical)
                    ? xCategories : processor.getCategories('x');

            let dataUpdate;
            if (datatypeX == ORDINAL_DATATYPE || datatypeY == ORDINAL_DATATYPE) {
                dataUpdate = processor.updateSeriesdatabyCategories(d3_values(data), xCategories, yCategories, yaxiscolumnorder);
            } else {
                dataUpdate = data;
            }

            if (chart.renderVisiblePointsOnly) {
                dataUpdate = dataProcessorHelpers.filterPointsInViewport(chart, dataUpdate, 'scatter', 'xy', this.builderC.datasetIndex);
            }
            this.seriesdataAfterUpdate[subseriesIdx] = dataUpdate;
            this.oldRenderer.seriesdataAfterUpdate[subseriesIdx] = dataUpdate; //update in existing renderer
            this.builderC.xcategoriesReversed = xCategories;
            if (this.builderC.xaxisreversed) {
                dataUpdate = dataUpdate.slice(0).reverse();
                this.builderC.xcategoriesReversed = xCategories ? xCategories.slice(0).reverse() : xCategories;
            }

            return this.seriesdataAfterUpdate[subseriesIdx];
        },
        position: {},
        styles: {},
        positionComponent: {},
        styleComponent: {}
    };
    return helpers;
};
