// @zohocharts/zc/src/scatter/parser/scatter-parser.js
import { renderDom } from "../../mark/dom-render";
import { executeMarkOperators } from "../../plot/mark-operator-executor";
import { appendEle, createElement } from "../../utils/zcsvgutils";
import { pick } from "../../utils/zcutils";
import { setScatterBuilderConfig } from "./config/scatter-builder-config";
import { scatter_encode_collectors } from "./encode/scatter-encode-helpers";
import { ZC_SCATTER_MARKS_MAPPING } from "./marks/scatter-marks-mapping";
import { scatter_parser_helpers } from "./scatter-parser-helpers";
import { getScatterCommonOption, setScatterRendererConfig } from "./config/scatter-renderer-config";

export class ScatterRenderer {
    constructor(args) {
        let { data, config, rendererIndex, legendOrder, paneid, chart, builderC, scatterDatasetIndex, phase, type, viewFlow } = args;
        this.data = data;
        this.config = config;
        this.paneid = paneid;
        this.rendererIndex = rendererIndex;
        this.scatterDatasetIndex = scatterDatasetIndex;
        this.conf = chart.rendererConf[rendererIndex];
        this.legendOrder = legendOrder;
        this.builderC = builderC;
        this.chart = chart;
        this.viewFlow = viewFlow || chart?.viewFlow;
        this.type = type;
        this.phase = phase;
        this.oldRenderer = args.oldRenderer;
        this.markOperator = args.operator;
        this.setMarkMapping();
    }

    initialize(container) {
        this.container = container;
        this.commonOption = getScatterCommonOption(this.conf, this.rendererIndex, this.legendOrder, this.paneid);
        setScatterBuilderConfig.call(this);     // transition and animation
        setScatterRendererConfig.call(this);    // rendering dependencies
        this.seriesdataAfterUpdate = new Array(0);
        this.oldRenderer.seriesdataAfterUpdate = new Array(0);
        this.scales = {
            x: this.commonOption.bandwidth1,
            interX: this.commonOption.bandwidth2,
            y: this.commonOption.yscale
        };
        this.helpers = scatter_parser_helpers.call(this);
        this.collectEncode();
    }
    collectEncode() {
        scatter_encode_collectors.call(this);
    }

    getAllowedMarkOperator() {
        return ['scatter'];
    }

    setMarkMapping(mapping) {
        this.markMapping = (mapping || ZC_SCATTER_MARKS_MAPPING).call(this);
    }

    getDefaultMarkParser() {
        let operators = this.getAllowedMarkOperator(),
            markMapping = this.markMapping,
            customOperator = this.markOperator;
        this._mark = executeMarkOperators.call(this, this.data, markMapping, operators, customOperator);
    }

    render(parent) {
        const { dynamicSeriesPosition, datasetIndex, renderingOrder, stacked, stackIndex, yrangePos } = this.builderC;
        let holderIndex = pick(dynamicSeriesPosition, datasetIndex),
            data_bind = { data: this.data, index: renderingOrder, stackIndex };

        let insertBefore;
        if (stacked) {
            // For ordering series to fix overlapping of markers
            let scatterseries_n = this.container.node().querySelector(`g.scatterseries_${dynamicSeriesPosition - 1}`);
            insertBefore = scatterseries_n ? `g.${scatterseries_n.className.baseVal} + *` : null;
        }

        this.placeHolder = createElement('g', `.scatterseries_${holderIndex}`, [data_bind], parent, insertBefore);

        if (this.type === 'arearange') {
            this.placeHolder = appendEle(this.placeHolder, 'g', [data_bind], '.', 'class', `rangePos_${yrangePos}`);
        }
        this._mark.map(d => {
            renderDom(d.mark, d.holder || this.placeHolder, this.viewFlow);
        });
    }

    destroy() {
        this.container = this.placeHolder = this._mark = null;
    }
}
