// @zohocharts/zc/src/scatter/bubbleOptions.js
import { defaultTheme as options } from '../core/defaulttheme';

options.bubble = function() {
    return {
        outerPadding: 0.02,
        multiColoring: false,
        marker: {
            innerFillOpacity: 0,
            innerStrokeWidth: 0,
            innerStrokeOpacity: 0,
            outerFillOpacity: 0.4,
            outerStrokeOpacity: 1
        },
        hoveroptions: {
            innerFillColor: 'transparent',
            innerStrokeColor: 'transparent',
            innerStrokeOpacity: 1,
            innerStrokeWidth: 0,
            outerFillOpacity: 0.3,
            outerStrokeOpacity: 0.5
        },
        datalabels: {
            showAs: 'y' //x,y,z,percent
        },
        animation: {
            type: 'sizing' //sizing || onebyone || seriesbyseries || allseries || fade || fadeAll || verticalAll
        },
        gradients: {
            // radial || linear || none
            options: {
                linear: {
                    y2: 100,
                    gradientUnits: 'userSpaceOnUse', //objectBoundingBox || userSpaceOnUse
                    colorGamma: [1.5, 0]
                },
                radial: {
                    radius: 90,
                    colorGamma: [1.5, 0]
                }
            }
        }
    };
};

export var _bubbleOptions = options.bubble;
