// @zohocharts/zc/src/scatter/bubbleSelection.js
import { pointInRectangle } from '../events/events-utils';
import { defined } from '../utils/zcutils';
import { d3_ascending } from '../core/d3-constants';
import { mathSQRT } from '../core/zc-constants';
import { ScatterSelection } from './scatterSelection';

export function bubbleSelection(points, mouse, chart, key) {
    var doRadiusSearch = chart.cache.search.bisect,
        points_scatter = doRadiusSearch ? ScatterSelection(points, mouse, chart, key) : points;
    var dataObject = chart.dataObject,
        plotArea = chart.plotarea,
        isComboChart = dataObject.isCombinationalChart,
        point,
        distance = function(x, y, cx, cy) {
            return mathSQRT((x - cx) * (x - cx) + (y - cy) * (y - cy));
        },
        overPoints;
    if(!points_scatter){
        return;
    }
    /**
     *
     * @type {{over: boolean, pointDistance: *, radius: number, point: *, bubbleDistance: number}[]}
     */
    var pointDetails = points_scatter
        .map(function(d) {
            var p = d._data(),
                rad = p._size / 2,
                xy = { x: p._x, y: p._y },
                dist = distance(mouse.x, mouse.y, xy.x, xy.y);

            // #ZC1805
            if (
                pointInRectangle(xy.x, xy.y, {
                    x: (chart.scrolledX || 0),
                    y: (chart.scrolledY || 0),
                    width: plotArea.width,
                    height: plotArea.height
                })
            ) {
                return {
                    radius: rad,
                    bubbleDistance: dist - rad,
                    pointDistance: dist,
                    over: dist - rad <= 0,
                    point: d
                };
            }
        })
        .filter(p => defined(p));
    /**
     * 1. If only one point has been found within the radius, the use the it directly without any further checking.
     */
    if (pointDetails.length === 1) {
        /**
         * For combo charts, marking it active only if it is exactly over the point
         */
        if (!isComboChart || (isComboChart && pointDetails[0].over)) {
            point = pointDetails[0].point;
        }
    } else if (pointDetails.length > 1) {
        // filter all the that's hovered.
        overPoints = pointDetails.filter(function(p) {
            return p.over;
        });

        /**
         * 2. If the mouse is hovered inside a single point, use the it directly without any further checking.
         */
        if (overPoints.length === 1) {
            point = overPoints[0].point;
        } else if (overPoints.length > 1) {
            /**
             * 3. If the mouse is hovered inside of intersection of multiple points, then
             *      1. We sort the intersected points by it's real distance from the mouse in ascending order.
             *      2. Use the closet point. (Which is the first point, since it is sorted in ascending order)
             */
            var overlappingFindBy = 'radius'; // pointDistance  | radius
            var pointSortedPointDistance = overPoints.slice(0).sort(function(a, b) {
                return d3_ascending(a[overlappingFindBy], b[overlappingFindBy]);
            });
            point = pointSortedPointDistance[0].point;
        } else {
            /**
             * 4. If the mouse is not inside any points, then we find the closest one.
             *      1. We sort the points within the tracking radius by it's bubble distance from the mouse in ascending order.
             *      2. Use the closet point. (Which is the first point, since it is sorted in ascending order)
             */
            var pointSortedDistance = pointDetails.slice(0).sort(function(a, b) {
                return d3_ascending(a.bubbleDistance, b.bubbleDistance);
            });
            point = pointSortedDistance[0].point;
        }
    }
    return point;
}
