// @zohocharts/zc/src/scatter/canvasScatterchart.js
//$Id$
//$Id$
//canvasScatterchart.js
import { DataProcessor } from '../core/data-processor';
import { mergeJSON, isUndefined, pick, d3_values, defined, invokeFunction} from '../utils/zcutils';
import { colorscale_helpers } from '../axis/colorscale-helpers';
import { isMultiColoring, getFillColor, getStrokeColor, rgba_canvas, getRawColor } from '../utils/colorutils';
import { getDashArray, getLineCap } from '../utils/dashstyles';
import { charttype, legend_helpers } from '../core/main-zc-init';
import { ORDINAL_DATATYPE } from '../core/datatype-constants';
import { getSegmentsForSymbol, getSVGsymbol } from '../utils/svg-symbols';
// import { legend_helpers } from '../legend/legend-helper';
import { effects } from '../effects/zceffects';
import { mathPI, NULL, DEVICE_PIXEL_RATIO } from '../core/zc-constants';
import { canvasUtils } from '../utils/canvasutils';
import { labelRenderer } from '../core/main-zc-init';
import { d3_rgb } from '../core/d3-constants';
import { START_ALIGN, MIDDLE_ALIGN, END_ALIGN } from '../core/align-constants';
import { isScatterChartType } from '../events/events-utils';
import { chartcategory, canvas_plot_renderer } from '../core/main-zc-init';
import { BubbleHighlighter } from './bubbleHighlight';
import { ScatterHighlighter } from './scatterHighlight';
import { getChartTypeByIndex } from '../core/charttype-helpers';
import { isFacet } from '../axis/hierarchy-axis-utils';
import { isObject } from '../utils/zcutils';

canvas_plot_renderer.scatter = function() {
    var shiftInfo,
        commonRendererProp,
        yrangePos,
        dataProcessorHelpers = DataProcessor.helpers;

    function chart(context, _s, pattern, filterIndex, encode = 'clr') {
        var rp = commonRendererProp.params,
            chartObj = rp.chartObj,
            xscale = rp.xscale,
            yscale = rp.yscale,
            stacked = rp.stacked,
            index = rp.datasetIndex,
            legendOrder = rp.legendOrderIndex,
            rendererIndex = rp.rendererIndex,
            axisrotated = rp.axisrotated,
            commonPlotOptions = rp.commonPlotOptions,
            color = rp.color,
            labelQueue = rp.labelQueue,
            yaxiscolumnorder = rp.yaxiscolumnorder,
            dataset = rp.dataset,
            threshold = rp.threshold,
            { shapescale } = rp,
            seriesData = chartObj.dataObject.categoryHierarchy ? _s[index] : chartObj.seriesdata[index],
            isSelectionContainer = !(chartObj.container.select('.selection-canvas').empty()),
            isSelectionCanvas = context.canvas.className == 'selection-canvas' ? true : false,
            selectionArray = chartObj.cache.clickSelectionData;

        var plotoptions = seriesData.plotoptions || {},
            cpOptions = plotoptions ? mergeJSON(commonPlotOptions, plotoptions, true) : commonPlotOptions,
            currentseriesdata = seriesData.data,
            paneid = isFacet(chartObj) ? `${seriesData.row}_${seriesData.column}` : '0_0',
            temp = getChartTypeByIndex(chartObj, paneid, index),
            cType = temp.value,
            chartName = temp.name,
            scatterOrBubble = isScatterChartType(cType),
            isBubbleChart = chartcategory.bubble.has(cType),
            isPolar = chartObj.dataObject.isPolarAxisCategory,
            legend_data = chartObj.systemConf.legend;

        var markerOptions = (commonRendererProp.params.markerOptions =
            plotoptions && plotoptions.marker ? mergeJSON(commonPlotOptions.marker, plotoptions.marker, true) : commonPlotOptions.marker);

        var hoverOptions = (commonRendererProp.params.hoverOptions =
                plotoptions && plotoptions.hoveroptions
                    ? mergeJSON(commonPlotOptions.hoveroptions, plotoptions.hoveroptions, true)
                    : commonPlotOptions.hoveroptions),
            symboltype = markerOptions.symbol || 'circle',
            symbolsize = markerOptions.size,
            imageUrl = markerOptions.imageUrl,
            useSVGMarkers = markerOptions.useSVGMarkers,
            // sizeRatio = markerOptions.sizeRatio ?? 0.5, //* change to markerOptions.sizeRatio ?? 0.5
            multiColoring = scatterOrBubble && isMultiColoring(cpOptions, chartObj),
            patternPallete = legend_data.patternPallete,
            hasColorScale = colorscale_helpers.isQuantileColorScale(chartObj);

        if (!scatterOrBubble) {
            if (isUndefined(symbolsize)) {
                symbolsize = pick(cpOptions.strokeWidth, 2);
            }

            if (!imageUrl) {
                symbolsize *= isPolar ? 0.56 : 0.7;
            }

            color = pick(seriesData.color, color);
        } else {
            symbolsize = pick(symbolsize, 1);
        }

        var _color = isFacet(chartObj) ? getRawColor(chartObj, legendOrder, seriesData, legendOrder, paneid) : color,
            outerSize = symbolsize * 5,
            outerFillOpacity = markerOptions.outerFillOpacity,
            outerStrokeOpacity = markerOptions.outerStrokeOpacity,
            innerFillOpacity = markerOptions.innerFillOpacity,
            innerStrokeOpacity = markerOptions.innerStrokeOpacity,
            outerStrokeWidth = pick(markerOptions.outerStrokeWidth, 1),
            innerStrokeWidth = pick(markerOptions.innerStrokeWidth, 1),
            outerStrokeColor = pick(invokeFunction(markerOptions.outerStrokeColor,_color,markerOptions,chartObj), _color),
            innerStrokeColor = pick(invokeFunction(markerOptions.innerStrokeColor,_color,markerOptions,chartObj), _color),
            outerFillColor = pick(invokeFunction(markerOptions.outerFillColor,_color,markerOptions,chartObj), _color),
            innerFillColor = pick(invokeFunction(markerOptions.innerFillColor,_color,markerOptions,chartObj), _color),
            outerDashArray = getDashArray(markerOptions.outerDashStyle, outerStrokeWidth),
            innerDashArray = getDashArray(markerOptions.innerDashStyle, innerStrokeWidth),
            outerStrokeLineCap = getLineCap(markerOptions.outerLineCap, markerOptions.outerDashStyle),
            innerStrokeLineCap = getLineCap(markerOptions.innerLineCap, markerOptions.innerDashStyle),
            markerPosition = markerOptions.position;

        if (defined(pattern)) {
            outerStrokeWidth = pick(pattern['stroke-width'], outerStrokeWidth);
           
            if (defined(filterIndex) && !(outerStrokeOpacity && d3_rgb(outerStrokeColor).opacity)) {
                innerStrokeWidth = pick(pattern['stroke-width'], innerStrokeWidth);
            }

            if(defined(pattern) && !isObject(pattern)){//for selection
                outerStrokeColor = (outerStrokeColor == '#ffffff00' || outerStrokeColor == 'transparent') ? outerStrokeColor : pick(pattern, outerStrokeColor);
                innerStrokeColor = (innerStrokeColor == '#ffffff00' || innerStrokeColor == 'transparent') ? innerStrokeColor : pick(pattern, innerStrokeColor);
                innerFillColor = (innerFillColor == '#ffffff00' || innerFillColor ==  'transparent') ? innerFillColor : pick(pattern, innerFillColor);
                outerFillColor = (outerFillColor == '#ffffff00' || outerFillColor == 'transparent') ? outerFillColor : pick(pattern, outerFillColor);
            }
        }

        if (defined(outerFillOpacity) && outerFillColor) {
            outerFillColor = rgba_canvas(outerFillColor, outerFillOpacity);
        }
        if (defined(outerStrokeOpacity) && outerStrokeColor) {
            outerStrokeColor = rgba_canvas(outerStrokeColor, outerStrokeOpacity);
        }
        if (defined(innerFillOpacity) && innerFillColor) {
            innerFillColor = rgba_canvas(innerFillColor, innerFillOpacity);
        }
        if (defined(innerStrokeOpacity) && innerStrokeColor) {
            innerStrokeColor = rgba_canvas(innerStrokeColor, innerStrokeOpacity);
        }

        function getSymbolSize(d, type = 'outer') {
            if (isPolar && (!d || d.isDummy || (useSVGMarkers === false && (isCatNullX(dataset.getX(d)) || isCatNullY(dataset.getY(d, yaxiscolumnorder)))))) {
                return 0;
            } else if (
                !isPolar &&
                cType !== cType.stkdarea &&
                (isCatNullX(dataset.getX(d)) || isCatNullY(dataset.getY(d, yaxiscolumnorder)))
            ) {
                return 0;
            } else if (!isPolar && cType == cType.stkdarea && (!d || d.isDummy)) {
                return 0;
            } else {
                return isBubbleChart ? chartObj.dataObject.getXYvalue(axisrotated, stacked, xscale, yscale, yaxiscolumnorder, d, 'z', chartObj.zscale, yrangePos)
                    : outerSize * 2;
                // let size = isBubbleChart ? chartObj.dataObject.getXYvalue(axisrotated, stacked, xscale, yscale, yaxiscolumnorder, d, 'z', chartObj.zscale, yrangePos)
                //     : outerSize;
                // type !== 'outer' ? (size *= sizeRatio) : 0;
                // return size;   //* sizeRatio code change
            }
        }

        function getSymbolBBox(d) {
            var size = getSymbolSize(d) / 2;
            if (size > 0 && (symboltype == 'line' || symboltype == 'vertical-line')) {
                if (symboltype == 'line') {
                    return {
                        width: size,
                        height: strokeWidth || 1
                    };
                } else if (symboltype == 'vertical-line') {
                    return {
                        width: strokeWidth || 1,
                        height: size
                    };
                }
            } else {
                return null;
            }
        }
        function getSegments(d) {
            var size = getSymbolSize(d);
            if (symboltype == 'line' || symboltype == 'vertical-line') {
                var _data = d._data();
                var symbolSegment = getSegmentsForSymbol(symboltype, size, _data._x, _data._y);
                if (symboltype == 'line' || symboltype == 'vertical-line') {
                    return [symbolSegment];
                } else {
                    return symbolSegment;
                }
            }
        }
        if (scatterOrBubble) {
            chart.highlighter = isBubbleChart ? new BubbleHighlighter(chartObj, chart) : new ScatterHighlighter(chartObj, chart);
        }
        //Assign renderer functions
        chart.getSymbolSize = getSymbolSize;
        chart.getSymbolBBox = getSymbolBBox;
        chart.getSegments = getSegments;

        // don't use cache in case of bubble, multicolor and colorOverlay.
        var useCache =
                !defined(threshold) &&
                !multiColoring &&
                (chartName === 'scatter' || chartName === 'line' || chartName === 'area') &&
                !patternPallete &&
                !imageUrl &&
                !hasColorScale &&
                !shapescale &&
                !isSelectionCanvas,
            cache;

        var fillColor = (chart.fillColorFunc = getFillColor(chartObj, legendOrder, paneid, null, !scatterOrBubble)),
            strokeColor = getStrokeColor(chartObj, legendOrder, paneid),
            innerFillColorFunc = getFillColor(chartObj, legendOrder, paneid, null, true);

        var cacheCanvas;
        if (useCache) {
            fillStyle = chart.fillColorFunc(
                legendOrder,
                seriesData,
                null,
                outerFillColor,
                null,
                scatterOrBubble ? [outerSize, outerSize] : null,
                false
            );
            cache = canvas_plot_renderer.scatter.getCache(
                symboltype,
                outerSize,
                outerStrokeColor,
                outerStrokeWidth,
                fillStyle,
                innerStrokeColor,
                innerStrokeWidth,
                innerFillColor,
                markerOptions
            );
            cacheCanvas = cache.canvas;
            var adjust = cache.adjustPosition,
                drawWidth = cache.width;
            
        }
        var usePatternPallete = scatterOrBubble;

        if (!seriesData.disabled) {
            var datatypeX = chartObj.dataset.getDataType('x'),
                datatypeY = chartObj.dataset.getDataType('y', yaxiscolumnorder),
                isCatNullX = dataProcessorHelpers.isCatNull(dataset, 'x'),
                isCatNullY = dataProcessorHelpers.isCatNull(dataset, 'y', yaxiscolumnorder),
                xCategories = datatypeX == ORDINAL_DATATYPE ? xscale.domain() : null,
                yCategories = datatypeY == ORDINAL_DATATYPE ? yscale.domain() : null,
                categories =
                    datatypeX == ORDINAL_DATATYPE && xscale.domain != null && !defined(chartObj.dataObject.hierarchical)
                        ? xCategories
                        : chartObj.dataObject.getCategories('x'),
                fillStyle,
                strokeStyle;
            //Initiated in dataprocessor
            if (chartObj.dataObject.markerSize[chartName]) {
                var mSize = symboltype == 'line' || symboltype == 'vertical-line' ? strokeWidth || 1 : outerSize * 2;
                chartObj.dataObject.markerSize[chartName].add(mSize);
            }

            chart.seriesdataAfterUpdate = [];

            currentseriesdata.forEach(function(data, j) {
                var dataUpdate;
                if (datatypeX == ORDINAL_DATATYPE || datatypeY == ORDINAL_DATATYPE) {
                    dataUpdate = chartObj.dataObject.updateSeriesdatabyCategories(
                        d3_values(data),
                        xCategories,
                        yCategories,
                        yaxiscolumnorder
                    );
                } else {
                    dataUpdate = data;
                }

                if (chartObj.renderVisiblePointsOnly) {
                    let filterAxisType = chartName === 'scatter' || chartName === 'bubble' ? 'xy' : 'x';
                    dataUpdate = DataProcessor.helpers.filterPointsInViewport(chartObj, dataUpdate, chartName, filterAxisType, index);
                }

                chart.seriesdataAfterUpdate[j] = dataUpdate;

                if (hasColorScale && defined(pattern) && encode === 'clr') {
                    dataUpdate = legend_helpers.filterSelectionByColor(dataUpdate, true, filterIndex, chartObj);
                }

                if (isBubbleChart && defined(pattern) && encode === 'z') {
                    dataUpdate = legend_helpers.filterSelectionByColor(dataUpdate, true, filterIndex, chartObj, 'z');
                }
                if(isSelectionContainer){
                    if(defined(selectionArray[0])){
                        chartObj.container.select('.zc-canvas').style("visibility", "hidden");
                    }
                    else{
                        chartObj.container.select('.selection-canvas').style("visibility", "hidden");
                        chartObj.container.select('.zc-canvas').style("visibility", "visible");
                    }
                }

                dataUpdate.forEach(function(d, i) {
                    var itemIndex = encode === 'z' ? pick(d.sortedItemIndex, i) : i,
                        outerFillStyle,
                        outerStrokeStyle,
                        position = i === 0 ? START_ALIGN : i === dataUpdate.length - 1 ? END_ALIGN : MIDDLE_ALIGN;

                    if (
                        !(
                            !isPolar &&
                            cType !== charttype.stkdarea &&
                            (!d || d.isDummy || isCatNullX(dataset.getX(d)) || isCatNullY(dataset.getY(d, yaxiscolumnorder)))
                        )
                    ) {
                        // to remove stacked area with dummy data added
                        if (!isPolar && cType == charttype.stkdarea && (!d || d.isDummy)) {
                            return;
                        }
                        if (markerPosition && markerPosition.indexOf(position) <= -1) {
                            return;
                        }

                        var x = chartObj.dataObject.getXYvalue(axisrotated, stacked, xscale, yscale, yaxiscolumnorder, d, 'x'),
                            y = chartObj.dataObject.getXYvalue(axisrotated, stacked, xscale, yscale, yaxiscolumnorder, d, 'y', null, yrangePos),
                            size = getSymbolSize(d) / 2,
                            shape = shapescale ? shapescale(d[dataset.shapeidx]) : symboltype;

                        if (useCache) {
                            // using cache image
                            context.beginPath();
                            var centered = size + outerStrokeWidth / 2 + adjust;
                            context.drawImage(cacheCanvas, x - centered, y - centered, drawWidth - 1, drawWidth - 1);
                            context.closePath();
                        } else {
                            var isSelectionUnselectedColor = isSelectionCanvas && !d.clickSelection,
                                selectionUnselectedColor = chartObj.systemConf.chart.plot.eventHandler.click.behaviour.unselected;
                            // outer circle
                            if (outerFillColor || outerStrokeColor) {
                                outerStrokeStyle = strokeColor(
                                    multiColoring ? itemIndex : legendOrder,
                                    d,
                                    chartObj.imagePallete ? null : invokeFunction(isSelectionUnselectedColor? selectionUnselectedColor : markerOptions.outerStrokeColor,_color,markerOptions,chartObj),
                                    NULL,
                                    outerStrokeOpacity
                                );
                                var drawScatterFunction = function(context, fillStyle, imageDimension) {
                                    canvas_plot_renderer.scatter.drawPath(
                                        shape,
                                        context,
                                        x,
                                        y,
                                        size,
                                        outerStrokeWidth,
                                        outerStrokeStyle,
                                        fillStyle,
                                        outerDashArray,
                                        outerStrokeLineCap,
                                        imageUrl,
                                        imageDimension,
                                        outerFillOpacity
                                    );
                                };

                                if (imageUrl) {
                                    var options = {
                                        x: x,
                                        y: y,
                                        width: size * 2,
                                        height: size * 2,
                                        context: context,
                                        name: imageUrl,
                                        repeatMode: 'no-repeat'
                                    };
                                    effects.canvas_patterns.imageFill(options, drawScatterFunction);
                                } else if (chartObj.imagePallete) {
                                    var options = {
                                        x: x,
                                        y: y,
                                        context: context,
                                        name: chartObj.imagePallete[multiColoring ? itemIndex : legendOrder]
                                    };
                                    effects.canvas_patterns.imageFill(options, drawScatterFunction);
                                } else {
                                    var fillStyle = chart.fillColorFunc(
                                        multiColoring ? itemIndex : legendOrder,
                                        d,
                                        null,
                                        invokeFunction(isSelectionUnselectedColor ? selectionUnselectedColor : markerOptions.outerFillColor,_color,markerOptions,chartObj),
                                        outerFillOpacity,
                                        scatterOrBubble ? [size, size] : null,
                                        false
                                    );
                                    drawScatterFunction(context, fillStyle);
                                }
                            }

                            // inner circle
                            var innerMarker = true,
                                isSelectionUnselectedColor = isSelectionCanvas && !d.clickSelection,
                                selectionUnselectedColor = chartObj.systemConf.chart.plot.eventHandler.click.behaviour.unselected;
                            if (
                                imageUrl ||
                                (innerStrokeWidth === 0 && (innerFillOpacity === 0 || invokeFunction(markerOptions.innerFillColor,_color,markerOptions,chartObj) === 'transparent'))
                            ) {
                                innerMarker = false;
                            }
                            // TODO gradients need to check since no such condition in svg renderer
                            // if (!(isBubbleChart && cpOptions.gradients.type && cpOptions.gradients.type !== NONE) && !imageUrl && (innerFillColor || innerStrokeColor)) {
                            if (innerMarker) {
                                fillStyle = innerFillColorFunc(
                                    multiColoring ? itemIndex : legendOrder,
                                    d,
                                    null,
                                    invokeFunction(isSelectionUnselectedColor ? selectionUnselectedColor : markerOptions.innerFillColor,_color,markerOptions,chartObj),
                                    innerFillOpacity,
                                    null,
                                    false
                                );
                                strokeStyle = rgba_canvas(
                                    strokeColor(
                                        multiColoring ? itemIndex : legendOrder,
                                        d,
                                        chartObj.imagePallete ? null : invokeFunction(isSelectionUnselectedColor ? selectionUnselectedColor : markerOptions.innerStrokeColor,_color,markerOptions,chartObj)
                                    ),
                                    innerStrokeOpacity
                                );
                                canvas_plot_renderer.scatter.drawPath(
                                    shape,
                                    context,
                                    x,
                                    y,
                                    size / 2,
                                    innerStrokeWidth,
                                    strokeStyle,
                                    fillStyle,
                                    innerDashArray,
                                    innerStrokeLineCap,
                                    null,
                                    null,
                                    null,
                                    'inner'
                                );
                            }
                        }
                    }
                });
            });
        }

        if (defined(labelRenderer)) {
            var markerSize;
            if (scatterOrBubble) {
                markerSize = outerSize * 2 + markerOptions.outerStrokeWidth;
            } else {
                var strokeWidth = pick(cpOptions.strokeWidth, 2);
                markerSize = pick(markerOptions.size, strokeWidth) * outerSize + strokeWidth / 2;
            }
            labelQueue.push({
                renderer: chart,
                charttype: 'scatter',
                chartObj: chartObj,
                rendererIndex: rendererIndex,
                paneid,
                args: [seriesData, legendOrder, markerSize, yrangePos]
            });
        }
        chart.chartRendered = true;
        context = cacheCanvas = null;   // memory leak fix
        return chart;
    }

    chart.getHighlighter = function(chartObj, index) {
        if (!chart.highlighter) {
            var isBubbleChart = chartcategory.bubble.has(getChartTypeByIndex(chartObj, null, index).value);
            chart.highlighter = isBubbleChart ? new BubbleHighlighter(chartObj, chart) : new ScatterHighlighter(chartObj, chart);
        }
        return chart.highlighter;
    };

    chart.commonRendererProp = function(_) {
        if (!arguments.length) {
            return commonRendererProp;
        }
        commonRendererProp = _;
        return chart;
    };
    chart.shiftInfo = function(_) {
        if (!arguments.length) {
            return shiftInfo;
        }
        shiftInfo = _;
        return chart;
    };

    chart.yrangePos = function(_) {
        if (!arguments.length) {
            return yrangePos;
        }
        yrangePos = _;
        return chart;
    };

    return chart;
};

/**
 * @param {String} symbol
 * @param {CanvasRenderingContext2D} context
 * @param {Number} x
 * @param {Number} y
 * @param {Number} size
 * @param {Number} strokeWidth
 * @param {String} strokeColor
 * @param {String} fillColor
 */

canvas_plot_renderer.scatter.drawPath = function(
    symbol,
    context,
    x,
    y,
    size,
    strokeWidth,
    strokeColor,
    fillColor,
    strokeDashArray,
    strokeLineCap,
    imageUrl,
    imageDimension,
    opacity,
    side
) {
    context.save();
    context.translate(x, y);
    context.beginPath();
    context.lineWidth = strokeWidth;
    context.strokeStyle = strokeColor;

    if (symbol === 'circle') {
        context.arc(0, 0, size, 0, mathPI * 2);
    } else {
        var path = getSVGsymbol(symbol)(size * 2, null, null, side);
        canvasUtils.drawPath(context, path);
    }

    if (imageUrl) {
        // For marker with imageUrl
        context.globalAlpha = opacity; //Set outerFillOpacity for imageUrl
        context.translate(-size + imageDimension.x, -size + imageDimension.y);
    }

    context.fillStyle = fillColor;
    if (fillColor) {
        context.fill();
    }

    if (imageUrl) {
        context.globalAlpha = 1; //Reset outerFillOpacity for imageUrl
    }

    if (strokeDashArray && strokeDashArray.length && typeof strokeDashArray === 'object' && defined(context.setLineDash)) {
        context.setLineDash(strokeDashArray);
        context.lineCap = strokeLineCap;
    }

    if (strokeWidth && strokeColor) {
        context.stroke();
    }
    context.closePath();
    context.restore();
};

/**
 * osc, osw, ofc, isc, isw, ifc - params
 * o - outer
 * i - inner
 * s - stroke
 * f - fill
 * c - color
 * w - width
 * ----
 * @param {String} symbol
 * @param {Number} outerSize
 * @param {String} osc
 * @param {Number} osw
 * @param {String} ofc
 * @param {String} isc
 * @param {Number} isw
 * @param {String} ifc
 */

canvas_plot_renderer.scatter.getCache = function(symbol, outerSize, osc, osw, ofc, isc, isw, ifc, markerOptions) {
    var adjustPosition = 10, // to remove the sharp edge
        scaleRatio = DEVICE_PIXEL_RATIO * 1.5,
        cacheCanvas = document.createElement('canvas'),
        tempPos = (outerSize + osw / 2) * scaleRatio + adjustPosition / 2,
        cache = cacheCanvas.getContext('2d'),
        canvasDimension = (outerSize * 2 + osw + adjustPosition) * scaleRatio,
        width = canvasDimension / scaleRatio;
    cacheCanvas.height = cacheCanvas.width = canvasDimension;
    // outer symbol
    this.drawPath(
        symbol,
        cache,
        tempPos,
        tempPos,
        outerSize * scaleRatio,
        osw * scaleRatio,
        osc,
        ofc,
        getDashArray(markerOptions.outerDashStyle, osw, scaleRatio),
        getLineCap(markerOptions.outerLineCap, markerOptions.outerDashStyle)
    );
    // inner symbol
    this.drawPath(
        symbol,
        cache,
        tempPos,
        tempPos,
        (outerSize / 2) * scaleRatio,
        isw * scaleRatio,
        isc,
        ifc,
        getDashArray(markerOptions.innerDashStyle, isw, scaleRatio),
        getLineCap(markerOptions.innerLineCap, markerOptions.innerDashStyle),
        null,
        null,
        null,
        
        'inner'
    );

    return {
        canvas: cache.canvas,
        width: width,
        adjustPosition: adjustPosition / 2 / scaleRatio
    };
};

export var _canvasScatterChart = canvas_plot_renderer.scatter;
