// @zohocharts/zc/src/scatter/scatterchart.js
//$Id$
//scatterchart.js

import { svg_plot_renderer } from '../core/main-zc-init';
import { utils, validateNullVal } from '../utils/zcutils';
import { d3_select } from '../core/d3-constants';
import { labelRenderer } from '../core/main-zc-init';
import { chartcategory } from '../core/main-zc-init';
import { ScatterHighlighter } from './scatterHighlight';
import { BubbleHighlighter } from './bubbleHighlight';
import { getChartTypeByIndex } from '../core/charttype-helpers';
import { isFacet } from '../axis/hierarchy-axis-utils';
import { ScatterRenderer } from './parser/scatter-parser';
import { BubbleRenderer } from './parser/extends/bubble-parser';
import { LineMarkerRenderer } from './parser/extends/line-marker-parser';
import { AreaMarkerRenderer } from './parser/extends/area-marker-parser';
import { GeoScatterRenderer } from './parser/extends/geoscatter-parser';
import { GeoBubbleRenderer } from './parser/extends/geobubble-parser';

/**
 *
 * @returns {chart} scatter, bubble renderer.
 */
svg_plot_renderer.scatter = function () {
    var chartObj,
        index,
        labelQueue,
        commonPlotOptions,
        redraw = false,
        transitionDuration;
    var shiftInfo,
        commonRendererProp,
        yrangePos,
        scatterDatasetIndex,
        newScatterParser = false,
        newScatterRenderer;
    function chart(selection) {
        selection.each(function (data) {
            //renderer config values
            const rp = commonRendererProp.params,
                legendOrder = rp.legendOrderIndex,
                rendererIndex = rp.rendererIndex;

            chartObj = rp.chartObj
            index = rp.datasetIndex
            commonPlotOptions = rp.commonPlotOptions
            labelQueue = rp.labelQueue

            var container = d3_select(this);
            chart.update = () => {
                d3_select(this).transition().call(chart);
            };

            var parser_data, data1;
            if (chartObj.viewFlow && chartObj.viewFlow.marks._default.chart.length) {
                parser_data = data[index];
                data1 = parser_data.data;
            }

            var currentdata = data1 || data[index],
                plotoptions = currentdata.plotoptions,
                plotConfig = chartObj.dataSplicer.getPlotConfig(parser_data || currentdata, index),
                gvp = (rp.gvp = utils.getValueOfPlotOptions(plotConfig, commonPlotOptions));

            var paneid = isFacet(chartObj) ? `${currentdata.row}_${currentdata.column}` : '0_0';

            var cType_Obj = getChartTypeByIndex(chartObj, paneid, index),
                chartName = cType_Obj.name;

            /* Try scatter parser here */
            newScatterParser = true;
            scatterDatasetIndex = index;
            let args = {
                type: chartName,
                chart: chartObj,
                oldRenderer: chart,
                data: data1 || data[index],
                operator: parser_data ? parser_data.operator : null,
                paneid, legendOrder, rendererIndex, scatterDatasetIndex,
                config: {
                    common: commonPlotOptions,
                    series: plotConfig,
                    merged: gvp
                },
                builderC: rp,
                phase: redraw ? 'redraw' : 'draw'
            };

            let scatterRenderer = getRenderer(args, chartName); // TODO: null -> other chartTypes like bubble
            commonRendererProp.parser = newScatterRenderer = scatterRenderer;
            scatterRenderer.initialize(container);

            scatterRenderer.getDefaultMarkParser();
            scatterRenderer.render(container);

            let { animationOptions, markerOptions } = scatterRenderer.builderC,
                outersize = markerOptions.outersize,
                isBubbleChart = scatterRenderer.commonOption.isBubbleChart,
                scatterOrBubble = scatterRenderer.commonOption.scatterOrBubble;

            transitionDuration = animationOptions.duration;

            if (scatterOrBubble) {
                chart.highlighter = isBubbleChart ? new BubbleHighlighter(chartObj, chart) : new ScatterHighlighter(chartObj, chart);
            }

            if (labelRenderer != null) {
                var markerSize;
                if (scatterOrBubble || isBubbleChart) {
                    markerSize = outersize + markerOptions.outerStrokeWidth;
                } else {
                    var strokeWidth = validateNullVal(utils.getValueOfPlotOptions(plotoptions, commonPlotOptions)('strokeWidth'), 2);
                    markerSize = (markerOptions.size || strokeWidth) * outersize * 0.5 + strokeWidth / 2;
                }
                labelQueue.push({
                    renderer: chart,
                    charttype: 'scatter',
                    chartObj: chartObj,
                    rendererIndex: rendererIndex,
                    paneid: paneid,
                    args: [currentdata, index, markerSize, yrangePos]
                });
            }

            let chartInstance = commonRendererProp.params.chartObj;
            chartInstance.timerObj[chartInstance.timerObj.length] = chart.renderedTimer = setTimeout(function () {
                redraw = true;
                chart.chartRendered = true;
            }, transitionDuration);

            function getRenderer(args, chartName) {
                switch (chartName) {
                    case 'bubble':
                        return new BubbleRenderer(args)
                    case 'line': 
                        return new LineMarkerRenderer(args)
                    case 'area':
                    case 'arearange':
                        return new AreaMarkerRenderer(args)
                    case 'geoscatter':
                        return new GeoScatterRenderer(args)
                    case 'geobubble': 
                        return new GeoBubbleRenderer(args)
                    default:
                        return new ScatterRenderer(args)
                } 
            }

            // container = null;  // memory leak
        });
        return chart;
    }
    chart.getHighlighter = function (chartObj, index) {
        if (!chart.highlighter) {
            var isBubbleChart = chartcategory.bubble.has(getChartTypeByIndex(chartObj, null, index).value);
            chart.highlighter = isBubbleChart ? new BubbleHighlighter(chartObj, chart) : new ScatterHighlighter(chartObj, chart);
        }
        return chart.highlighter;
    };
    chart.commonRendererProp = function (_) {
        return arguments.length ? (commonRendererProp = _, chart) : commonRendererProp;
    };
    chart.shiftInfo = function (_) {
        return arguments.length ? (shiftInfo = _, chart) : shiftInfo;
    };

    chart.yrangePos = function (_) {
        return arguments.length ? (yrangePos = _, chart) : yrangePos;
    };
    chart.destroy = function (_) {
        newScatterRenderer && newScatterRenderer.destroy();
    }
    return chart;
};

export var _scatterChart = svg_plot_renderer.scatter;
