// @zohocharts/zc/src/scatter/scatterHighlight.js
import { NONE, NULL, TRANSPARENT } from '../core/zc-constants';
import { getPropVal, invokeFunction, simpleClone } from '../utils/zcutils';
import { addSVGmarker } from '../utils/svg-symbols';
import { getRawColor, isMultiColoring } from '../utils/colorutils';
import { HIGHLIGHT_CONTEXT, SELECTION_CONTEXT } from '../events/visualEffects/highlight/highlight-context';
import { appendEle } from '../utils/zcsvgutils';
// import { legend_helpers as L_H } from '../legend/legend-helper';
import { isBubbleChartType, isPointVisible } from '../events/events-utils';
import { charttype, HIGHLIGHT_EFFECT_REGEX, legend_helpers as L_H } from '../core/main-zc-init';
import { LOCATION_MARKERS } from '../geo-core/geo-marker-constant';
import { getChartTypeByIndex } from '../core/charttype-helpers';

export function ScatterHighlighter(chart, renderer) {
    this.chart = chart;
    this.renderer = renderer;
    return this;
}

var proto = ScatterHighlighter.prototype;

proto._defaultH = function() {
    return [
        {
            shape: 'marker',
            data: proto.data,
            path: this.path,
            context: this.renderingContext,
            translate: this.translateContext
        }
    ];
};

proto.data = function(point) {
    var rp = this.renderer.commonRendererProp().params;
    var chart = this.chart,
        renderer = this.renderer;

    return function() {
        if (!chart.cache.data[point._key]) {
            chart.cache.data[point._key] = {};
        }
        var cache = chart.cache.data[point._key];
        cache._x = point._x =
            cache._x || chart.dataObject.getXYvalue(rp.axisrotated, rp.stacked, rp.xscale, rp.yscale, rp.yaxiscolumnorder, point, 'x');
        cache._y = point._y =
            cache._y || chart.dataObject.getXYvalue(rp.axisrotated, rp.stacked, rp.xscale, rp.yscale, rp.yaxiscolumnorder, point, 'y');
        cache._size = point._size = cache._size || renderer.getSymbolSize(point);
        cache._visible = point._visible = isPointVisible(chart, point);
        cache._uid = point._uid = chart.dataset.getX(point);
        cache._point = point;
        return point;
    };
};

proto.commonOptions = function(point) {
    var chart = this.chart,
        renderer = this.renderer,
        rp = renderer.commonRendererProp().params,
        markerOptions = rp.markerOptions || {},
        hoverOptions = rp.hoverOptions || {};

    var symbol = hoverOptions.symbol || markerOptions.symbol,
        size = hoverOptions.size != null ? 1 : markerOptions.size; //CXZAEVENTS_COMBO_2
    // size = hoverOptions.size || markerOptions.size;

    var color = getRawColor(
        this.chart,
        isMultiColoring(rp.commonPlotOptions, this.chart) ? point.sortedItemIndex : point.seriesIndex,
        point,
        point.seriesIndex,
        point.paneid,
        point.rendererIndex
    );

    return {
        options: hoverOptions,
        symbol: symbol,
        _size: size,
        color: color,
        size: {
            outer: size,
            inner: size
        }
    };
};
proto.options = function(point) {
    var commonOptions = this.commonOptions(point),
        symbol = commonOptions.symbol,
        size = commonOptions._size || 1;

    var sizeFactor_O = LOCATION_MARKERS.has(symbol) ? 1 : 2, //TODO SOON MODULARIZE
        sizeFactor_I = LOCATION_MARKERS.has(symbol) ? 0.5 : 1;

    commonOptions.size = {
        outer: size * sizeFactor_O,
        inner: size * sizeFactor_I
    };

    return commonOptions;
};

proto.renderingContext = function(isSelection) {
    if(isSelection){
        return SELECTION_CONTEXT.get('selection').call(this, this.chart);
    }
    return HIGHLIGHT_CONTEXT.get('scatter').call(this, this.chart);
};

proto.translateContext = function() {
    return [this.chart.plotarea.left, this.chart.plotarea.top];
};

proto.path = function(point, model, isSelection) {
    point = (point._data && point._data()) || point;
    var _options = this.options(point),
        _hoveroptions = simpleClone(_options.options);
        _hoveroptions = hoverColorUpdater(_hoveroptions,_options,this.chart);

    const parser = this.renderer.commonRendererProp().parser,   // #ZC4684 - hoveroptions customization using scatter parser
        hoverComponent = getPropVal(parser, 'helpers.styleComponent.scatter.hoveroptions');
    hoverComponent ? (_hoveroptions = invokeFunction(hoverComponent, point, point.seriesIndex, { ..._hoveroptions })) : null;
        
    var context = model.context.call(this, isSelection),
        translate = model.translate.call(this),
        { seriesIndex, paneid } = point,
        cType = getChartTypeByIndex(this.chart, paneid, seriesIndex).value,
        isPointType = isBubbleChartType(cType) || cType === charttype.scatter,
        pathContext = appendEle(
            context,
            'g',
            [1],
            '.',
            'class',
            `highlight${point._key}${model.type}` || '_'
        ),
        getLegendHighlightEffect = this.chart.legendHighlight.getLegendHighlightEffect,
        highlightEffect = getLegendHighlightEffect && getLegendHighlightEffect(seriesIndex, 'selectedSeries', 'plot', paneid),
        isPatternPalleteEnabled =
            L_H.hasPatternPallete && L_H.hasPatternPallete(this.chart) &&
            highlightEffect !== 'strokeEnlarge' &&
            highlightEffect.constructor !== Function &&
            isPointType, //Dont apply pattern effects fro line charts and for scatter chart with custom callback effects and strokeEnlarge
        mPosition = model.type == 'end' ? [point._x, point._y0] : [point._x, point._y];
    addSVGmarker(
        pathContext,
        null,
        isPatternPalleteEnabled ? { symbol: _options.symbol } : _hoveroptions,
        {
            color: isPatternPalleteEnabled ? null : _options.color,
            symbolSize: {
                inner: _options.size.inner,
                outer: _options.size.outer
            },
            symbol: _options.symbol,
            transform: mPosition,
            imageSize: 15
        },
        this.chart
    );
    if (isPatternPalleteEnabled) {
        const getEffect = this.chart.legendHighlight.getEffect;
        var highlightStyle =
            highlightEffect === NONE
                ? {
                      fill: TRANSPARENT
                  }
                : getEffect && getEffect(
                      highlightEffect,
                      'scatter',
                      HIGHLIGHT_EFFECT_REGEX.test(highlightEffect) ? _options.color : TRANSPARENT,
                      { fill: TRANSPARENT, stroke: _options.color },
                      point.seriesIndex,
                      'plot'
                  );
        pathContext.styles(highlightStyle);
    }
    context.style('display', '');
    context.attr('transform', `translate(${translate[0]},${translate[1]})`);
    return pathContext;
};

function hoverColorUpdater(_hoveroptions,_options,chartObj){
    _hoveroptions.innerFillColor=invokeFunction(_hoveroptions.innerFillColor,_options.color,_options.options,chartObj);
    _hoveroptions.innerStrokeColor=invokeFunction(_hoveroptions.innerStrokeColor,_options.color,_options.options,chartObj);
    _hoveroptions.outerFillColor=invokeFunction(_hoveroptions.outerFillColor,_options.color,_options.options,chartObj);
    _hoveroptions.outerStrokeColor=invokeFunction(_hoveroptions.outerStrokeColor,_options.color,_options.options,chartObj);
    return _hoveroptions;
}
