// @zohocharts/zc/src/scatter/scatterOptions.js
import { defaultTheme as options } from '../core/defaulttheme';

options.scatter = function(chartType) {
    var options = {
        outerPadding: 0.02,
        multiColoring: false,
        marker: {
            size: 1,
            innerFillColor: 'white',
            innerStrokeColor: 'white',
            innerStrokeWidth: 0,
            innerStrokeOpacity: 1,
            outerFillColor: 'white',
            outerStrokeOpacity: 1,
            outerStrokeWidth: 2
        },
        hoveroptions: {
            innerStrokeOpacity: 1,
            outerFillColor: 'white',
            outerStrokeWidth: 2
        },
        datalabels: {
            showAs: 'y' //x,y,percent
        },
        animation: {
            type: 'allseries' //onebyone || seriesbyseries || allseries || fade || fadeAll
        },
        gradients: {
            // radial || linear || none
            options: {
                linear: {
                    y2: 100,
                    gradientUnits: 'userSpaceOnUse', //objectBoundingBox || userSpaceOnUse
                    colorGamma: [2, 0]
                },
                radial: {
                    radius: 90,
                    colorGamma: [2, 0]
                }
            }
        }
    };

    if (chartType === 'geoscatter') {
        // define the marker type
        var marker = options.marker,
            hoverOptions = options.hoveroptions;
        // marker.innerStrokeWidth = marker.innerFillOpacity = hoverOptions.innerFillOpacity = hoverOptions.innerStrokeWidth = 0;

        options.animation.type = 'sizing';
        options.datalabels.showAs = 'geo';
        options.datalabels.orient = 'bottom';
        options.datalabels.textOverlap = 'hidden';
    }

    return options;
};

export var _scatterOptions = options.scatter;
