// @zohocharts/zc/src/behaviours/listeners/brush/behaviour.js
import { d3_mouse } from '../../../core/d3/mouse';
import { each, invokeFunction, isArray } from '../../../utils/zcutils';
import { BrushD3 } from './brush-d3';
import { d3_event, d3_select } from '../../../core/d3-constants';

/**
 *
 * @param brush
 * @param selectionForce
 */
function positionSliderToMouseCenter(brush, selectionForce, event) {
    let type = brush.options.type;

    d3_event(event);
    const selection = selectionForce || brush.getSelection();
    const center = d3_mouse(this)[type === 'x' ? 0 : 1];
    const size = selection[1] - selection[0];

    const [x0, x1] = [center - size / 2, center + size / 2];
    const [minX, maxX] = brush.getRange();

    let newSelection = x1 > maxX ? [maxX - size, maxX] : x0 < minX ? [minX, minX + size] : [x0, x1];

    // call the move
    brush.setSelection(newSelection);
}

/**
 *
 * @param selection
 */
function fixEdgeCancellation(selection, brush) {
    // actually d3 cancels brush when start and end point are same.
    // when we move both the handle to the start or end position the start and end position become same and it get cancelled.
    // so we are adjust the position a little bit.

    if (selection[0] === selection[1]) {
        selection[1] += 0.000000001;
        selection[0] -= 0.000000001;
        brush.setSelection(selection, 0);
    }
}

function closestValue(array, value) {
    var result, lastDelta;

    array.some(function(item) {
        var delta = Math.abs(value - item);
        if (delta > lastDelta) {
            return true;
        }
        result = item;
        lastDelta = delta;
    });
    return result;
}

let brushBehaviours = {
    /**
     * Scroll behaviour
     */

    scroll: function() {
        let brush = this,
            last = 0,
            brushSelectionFn = null;
        this._track.datum({ type: 'selection' }).on('mousedown touchstart', function(event) {
            let selectionArgs = {
                event,
                scrollType: brush.options.type,
                brushInstance: brush,
                scrollScale: brush.options.brushScale,
                ismousedownTouchstart: true
            }
            brushSelectionFn = brush.options.brushSelection;
            if(brushSelectionFn != null && typeof brushSelectionFn === "string"){
                d3_event(event);
                invokeFunction(brushSelectionFn, selectionArgs);
            }else{
                positionSliderToMouseCenter.call(this, brush, null, event);
            }
        });

        brush.addBehaviourHook('brush.scroll', () => {
            last = brush.getSelection();
        });

        brush.addBehaviourHook('end.scroll', event => {
            if (!event.sourceEvent || last[0] !== last[1]) {
                return;
            }
            fixEdgeCancellation(last, brush);
        });
    },

    /**
     * move the current selected area to the mouse clicked
     * also support re-selection
     */

    reposition: function() {
        let brush = this,
            last;

        brush.addBehaviourHook('brush.reposition', () => (last = brush.getSelection()));
        brush.addBehaviourHook('end.reposition', (event, brushed) => {
            if (!event.sourceEvent) {
                return;
            }
            if (!brushed) {
                positionSliderToMouseCenter.call(brush.parent_group.node(), brush, last);
            }
        });
    },

    /**
     *
     * resize the brush selection to the mouse position
     */
    resizeToMouse: function() {
        let brush = this,
            last = brush.getSelection(),
            type = brush.options.type;

        brush.addBehaviourHook('brush.resizeToMouse', () => {
            last = brush.getSelection();
        });

        brush.addBehaviourHook('end.resizeToMouse', (event, brushed) => {
            const sourceEvent = event.sourceEvent;

            if (sourceEvent instanceof TouchEvent) {
                event = sourceEvent;
            } else if (!(sourceEvent instanceof MouseEvent)) {
                return;
            }

            if (!brushed && isArray(last)) {
                let position = d3_mouse(brush.parent_group.node(), event)[type === 'x' ? 0 : 1],
                    index = last.indexOf(closestValue(last, position));

                last[index] = position;
                brush.setSelection(last);
            } else {
                fixEdgeCancellation(last, brush);
            }
        });
    },

    slider: function() {
        let brush = this,
            last = brush.getSelection(),
            type = brush.options.type,
            resizeOnClick = (event) => {
                d3_event(event);
                let [x, y] = d3.pointer(event, brush.parent_group.node()),
                    val = type ? (type == 'x' ? x : y) : x;
                brush.setSelection([0, val]);
            },
            sliderRemoved = false;
        // stop resize
        this._selection.datum({ type: 'none' });
        this._track.datum({ type: 'none' });
        this._track.on('click touchstart', resizeOnClick);
        this._selection.on('click touchstart', resizeOnClick);

        brush.addBehaviourHook('start.slider', e => {
            if (!sliderRemoved) {
                brush.customHandles && d3_select(brush.customHandles.node()).remove();
            }
        });

        brush.addBehaviourHook('brush.slider', event => {
            let [x1, x2] = event.selection;
            last = [0, x2];
            if (x1 > 0) {
                brush.setSelection(last);
            }
        });
        brush.addBehaviourHook('end.slider', event => event.sourceEvent && fixEdgeCancellation(last, brush));
    }
};

// add the behaviours
each(brushBehaviours, (behaviour, name) => BrushD3.addBehaviour(name, behaviour));

export { brushBehaviours as _brushBehaviours };
