// @zohocharts/zc/src/behaviours/listeners/brush/brush-d3.js
import { d3_brush, d3_brushSelection, d3_brushX, d3_brushY, d3_dispatch, d3_event, d3_map, d3_select } from '../../../core/d3-constants';
import { boundToRect, defined, each, isArray, isNumber, pick, setFrozenVal, splat } from '../../../utils/zcutils';
import { createElement, setStyles } from '../../../utils/zcsvgutils';
import { isFunction } from '../../../utils/function-utils';
import { Registry } from '../../../core/zc-registry';
import { mathMax, mathMin } from '../../../core/zc-constants';

import * as customBrushMap from './custom-brush';
import { $Browser } from '../../../core/browser-detect';
const d3BrushMap = {
    brush: d3_brush,
    brushX: d3_brushX,
    brushY: d3_brushY
};

const defaultOptions = {
    // * For wheel zoom
    extent: [[0, 0], [0, 0]],
    handleSize: 6,
    type: 'xy',
    snap: null,
    element: null,
    metaKey: null,
    styles: {
        track: {},
        selection: {},
        handles: {}
    }
};

/**
 * This is used as a interface for zoom position
 */
export class BrushD3Options {
    constructor(options) {
        this.settings = { ...defaultOptions, ...options };
    }
}

/**
 *
 */
class BrushD3 {
    /**
     *
     * @param {BrushD3Options} brushD3Options
     */
    constructor(brushD3Options) {
        let options = (this.options = brushD3Options.settings);
        this.type = this.options.type;
        this.initialize();
        this.hookCount = 0;
        this._snap = {};

        if (defined(options.behaviour)) {
            if (BrushD3.behaviours.has(options.behaviour)) {
                this.behaviour = BrushD3.behaviours.get(options.behaviour);
            } else {
                throw new TypeError(`${options.behaviour} behaviour not found`);
            }
        }

        if (options.element && options.element.isD3Selection) {
            this.bindTo(options.element);
        }
    }

    initialize() {
        var brush = this,
            options = brush.options,
            useCustomBrush = $Browser.isTouchDevice && options.multitouch === true,
            extent = options.extent,
            type_map = {
                x: 'brushX',
                y: 'brushY',
                xy: 'brush'
            },
            brush_type = (brush.type = pick(type_map[brush.type], 'brush')),
            /**
             * @type BrushBehavior
             */
            d3_brush = (brush.d3_brush = brush.d3_brush || (useCustomBrush ? customBrushMap : d3BrushMap)[brush_type]()),
            signal,
            behaviour_signal;

        // set the extent if defined;
        if (isArray(extent) && isArray(extent[0]) && isArray(extent[1])) {
            d3_brush.extent(extent);
        }

        if (options.metaKey) {
            d3_brush.filter(function() {
                return !d3_event()?.[options.metaKey];
            });
        }

        // must call before the rendering
        d3_brush.handleSize(pick(options.handleSize, 6));

        // handle the events
        signal = brush._signal = d3_dispatch('start', 'brush', 'end');
        behaviour_signal = brush.behaviour_signal = d3_dispatch('init', 'start', 'brush', 'end');

        let lastSelection = [0, 0],
            brushed = false,
            started = false,
            mode;

        d3_brush.on('start', function(event){
            d3_event(event);
            // ignore when reset
            let sourceEvent,
                ee = event;
            if (!event.selection) {
                return false;
            }

            brush.brushed = brushed = false;
            started = false;

            while ((sourceEvent = ee.sourceEvent)) {
                ee = sourceEvent;
            }

            let sourceElement = ee?.target?.getAttribute ? ee.target : null;
            mode = sourceElement && /handle/.test(sourceElement.getAttribute('class')) ? 'resize' : 'move';

            setFrozenVal(event, 'mode', mode);
            behaviour_signal.call('start', brush, event);
            signal.call('start', brush, event);
        });

        d3_brush.on('brush', function(event){
            d3_event(event);
            // ignore when reset
            if (BrushD3.isInvalidBrush()) {
                return false;
            }

            // call the start on first brush mouse
            // if (!started) {
            //     // let e = d3_event();
            //     // e.type = 'start';
            //     // started = true;
            //     // signal.call('start', brush, e);
            // }
            setFrozenVal(event, 'mode', mode);

            let snapping = brush._snap.when === 'brush' && isFunction(brush._snap.calculate) && brush._snap.calculate(event);

            if (snapping) {
                if (BrushD3.isInvalidBrush()) {
                    return;
                }

                let snapSelection = pick(snapping, lastSelection);
                brush.setSelection(snapSelection);

                if (lastSelection[0] !== snapSelection[0] || lastSelection[1] !== snapSelection[1]) {
                    lastSelection = snapSelection;
                    setFrozenVal(event, 'selection', snapSelection);
                    brush.positionCustomHandles(lastSelection);
                    behaviour_signal.call('brush', brush, event);
                    signal.call('brush', brush, event);
                }
            } else {
                brush.positionCustomHandles();
                behaviour_signal.call('brush', brush, event);
                signal.call('brush', brush, event);
            }

            // detect manual brush
            if (event && event.sourceEvent && event.sourceEvent.type !== 'end') {
                brush.brushed = brushed = true;
            }
        });

        d3_brush.on('end', function(event){
            d3_event(event);
            if (BrushD3.isInvalidBrush()) {
                return;
            }

            let selection = brush.getSelection(),
                snapping = brush._snap.when === 'end' && isFunction(brush._snap.calculate) && brush._snap.calculate(brush, event);
            // set the mode
            setFrozenVal(event, 'mode', mode);

            if (snapping && selection) {
                let snapSelection = pick(snapping, lastSelection);
                brush.setSelection(snapSelection, brush._snap.duration);

                if (lastSelection[0] !== snapSelection[0] || lastSelection[1] !== snapSelection[1]) {
                    lastSelection = snapSelection;
                    setFrozenVal(event, 'selection', snapSelection);
                    behaviour_signal.call('end', brush, event, brushed);
                    signal.call('end', brush, event, brushed);
                }
            } else {
                brush.positionCustomHandles();
                behaviour_signal.call('end', brush, event, brushed);
                signal.call('end', brush, event, brushed);
            }
            d3_event(null);     //* resetting d3.event so that helps to create new BrushEvent in d3
        });
    }

    snap(when, calculator, duration) {
        this._snap = {
            when: when,
            calculate: calculator,
            duration: duration
        };
    }

    /**
     *
     * @return {{}}
     */
    getSignals() {
        return this._signal;
    }

    /**
     *
     * @param type
     * @param hook
     * @return {string}
     */
    addHook(type, hook) {
        this.hookCount++;
        let currentHookID = `${type}.${this.hookCount}`;
        this._signal.on(currentHookID, hook);
        return currentHookID;
    }

    /**
     *
     * @param currentHookID
     */
    removeHook(currentHookID) {
        this._signal.on(currentHookID, null);
    }

    /**
     *
     * @param type
     * @param hook
     * @return {string}
     */
    addBehaviourHook(type, hook) {
        this.hookCount++;
        let currentHookID = `${type}.${this.hookCount}`;
        this.behaviour_signal.on(currentHookID, hook);
        return currentHookID;
    }

    /**
     *
     * @param currentHookID
     */
    removeBehaviourHook(currentHookID) {
        this.behaviour_signal.on(currentHookID, null);
    }

    bindTo(parent) {
        let brush = this,
            d3_brush = brush.d3_brush,
            styles = brush.options.styles || {};

        // let group = createElement('g', '.brush-group-wrapper', [1], parent);
        let group = parent;
        // save the brush
        brush.parent_group = group;
        // call the brush
        group.call(d3_brush);

        // save the elements
        brush._track = group.selectAll('rect.overlay');
        brush._selection = group.selectAll('rect.selection');
        brush._handles = group.selectAll('rect.handle');

        // apply the styles
        setStyles(brush._track, styles.track);
        setStyles(brush._selection, styles.selection);
        setStyles(brush._handles, styles.handle);

        // if resizer is false, remove the nodes because, there is no API to remove them.
        if (brush.options.resizer === false) {
            group.selectAll('rect.handle').remove();
        }

        if (defined(brush.behaviour)) {
            brush.behaviour.call(brush);
        }

        // call the behaviour signal
        brush.behaviour_signal.call('init', brush);
    }

    addCustomHandle(handles) {
        let brush = this;

        // register external custom handle
        if (handles.isD3Selection) {
            return (brush.customHandles = handles);
        } else if (handles[0].shape === 'default') {
            return setStyles(brush._handles, handles[0].style);
        }

        brush.customHandles = BrushD3.ADD_HANDLE(handles, brush.parent_group).attrs({
            cursor: d => BrushD3.getCursor(d.type),
            'pointer-events': 'all'
        });

        brush.positionCustomHandles();

        return brush.customHandles;
    }

    /**
     *
     */
    positionCustomHandles() {
        let brush = this,
            selection = brush.getSelection();

        
        if (brush.customHandles) {
            if (selection) {
                    brush.customHandles
                    .attr('transform', function(d) {
                        let position = brush.getCustomHandlePosition(d);
                        return `translate(${position.join(' ')})`;
                    })
                    .style('display', null);
            } else {
                brush.customHandles.style('display', 'none');
            }
        }
    }

    /**
     *
     * @param handle
     * @return {*}
     */
    getCustomHandlePosition(handle) {
        let brush = this,
            selection = brush.getSelection(),
            brushArea = boundToRect(pick(brush.options.extent, brush.d3_brush.extent()())),
            alignCenter = handle.alignCenter !== false,
            centerX = 0,
            centerY = 0;
        // center the handle
        if (alignCenter) {
            centerX = brushArea.width / 2;
            centerY = brushArea.height / 2;
        }

        if (selection) {
            return {
                e: [selection[1], centerY],
                w: [selection[0], centerY],
                n: [centerX, selection[0]],
                s: [centerX, selection[1]]
            }[handle.type];
        }
    }

    /**
     *
     * @param selection
     * @param transitionDuration
     */
    setSelection(selection, transitionDuration, raceSettings) {
        let brush = this,
            type = brush.options.type,
            range = brush.getRange(type),
            isRaceSetSelection = raceSettings;

        if (this._raceInProgress && !isRaceSetSelection) {
            return;
            // we return for all cases other than isRacing, since if left for all cases, then transition will get piled up and we will be getting "too late already running" error.
        }
        this._raceInProgress = isRaceSetSelection ? true : false;
        
        // #ZC2166 constrain
        if (type === 'xy') {
            selection = [
                [mathMax(range[0][0], selection[0][0]), mathMin(range[0][1], selection[0][1])],
                [mathMin(range[1][0], selection[1][0]), mathMin(range[1][1], selection[1][1])]
            ];
        } else {
            selection = [mathMax(range[0], selection[0]), mathMin(range[1], selection[1])];
        }

        if (d3_event()) {
            setFrozenVal(d3_event(), 'selection', selection);
        }

        var raceTransition = raceSettings && raceSettings.syncTransition;
        if(raceTransition){
            brush.parent_group
            .transition(raceTransition)
            .call(brush.d3_brush.move, selection)
            .on('end', function() {
                raceSettings.visibleFrameStatus.isTransitionStarted = false;
            });
        }else{
            if (isNumber(transitionDuration) && transitionDuration > 0) {
                brush.parent_group
                    .transition()
                    .duration(transitionDuration)
                    .call(brush.d3_brush.move, selection);
            }else {
                brush.__checkSameType = true;
                brush.parent_group.call(brush.d3_brush.move, selection, d3_event());
            }
        }
        
    }

    getSelection() {
        return (d3_event()?.selection) || d3_brushSelection(this.parent_group.node());
    }

    /**
     *
     * @return {[]}
     */
    getRange() {
        let brush = this,
            extent = brush.d3_brush.extent()();

        switch (brush.options.type) {
            case 'x':
                return [extent[0][0], extent[1][0]];
            case 'y':
                return [extent[0][1], extent[1][1]];
            default:
                return extent;
        }
    }

    hide() {
        this.parent_group.attr('display', 'none');
    }

    show() {
        this.parent_group.attr('display', null);
    }

    reset() {
        this.parent_group.call(this.d3_brush.move, null);
    }

    destroy() {
        var brush = this;
        // #ZC2766
        brush.parent_group && brush.parent_group.on('.brush', null);
        brush.parent_group && brush.parent_group.selectAll('rect.overlay, rect.handle, rect.selection').remove();
        d3_event(null);

        each(brush, function(value, key) {
            brush[key] = null;
            delete brush[key];
        });
    }

    update(options) {
        this.options = options;
        this.type = pick(options.type, 'xy');
        this.initialize();
    }

    updateExtent(extent) {
        if (isArray(extent) && isArray(extent[0]) && isArray(extent[1])) {
            this.d3_brush.extent(extent);
            if (this.parent_group) {
                this.parent_group.call(this.d3_brush);
            }
        }
    }

    static isInvalidBrush(type) {
        let event = d3_event(),
            sourceEvent = event && event.sourceEvent;

        return sourceEvent && sourceEvent.type === (type || d3_event().type);
    }

    static getCursor(direction) {
        return {
            e: 'ew-resize',
            w: 'ew-resize',
            n: 'ns-resize',
            s: 'ns-resize'
        }[direction];
    }

    /**
     *
     * @param handles
     * @param parent
     */
    static ADD_HANDLE(handles, parent) {
        handles = splat(handles);

        let handle_selection = createElement('g', '.handle--custom', handles, parent);

        // create the shapes
        createElement(
            'path',
            '.custom-handle-path',
            (d, i) => {
                let shapes = splat(d.shape);

                if (isArray(d.shape[i])) {
                    shapes = d.shape[i];
                }

                return splat(shapes).map(shape => {
                    return { type: d.type, shape: shape, i: i };
                });
            },
            handle_selection
        )
            .attr('d', d => d.shape)
            .each(function(d, index) {
                let selection = d3_select(this),
                    handle = handles[d.i],
                    style = splat(handle.style),
                    attributes = splat(handle.attributes);
                // add the attributes if any
                selection.attrs(pick(attributes[index], handle.attributes));
                // set the styles
                setStyles(selection, pick(style[index], handle.style));
            });

        return handle_selection;
    }
}

BrushD3.prototype.on = BrushD3.prototype.addHook;

/**
 *
 */
BrushD3.behaviours = d3_map();
/**
 *
 * @param name
 * @param behaviour
 * @return {*}
 */
BrushD3.addBehaviour = function(name, behaviour) {
    if (!defined(behaviour)) {
        return BrushD3.behaviours.remove(name);
    }
    BrushD3.behaviours.set(name, behaviour);
};

export { BrushD3 };

Registry.setComponent('BrushD3', BrushD3);
Registry.setComponent('BrushD3Options', BrushD3Options);
