// @zohocharts/zc/src/behaviours/listeners/brush/custom-brush.js
import { createElement } from '../../../utils/zcsvgutils';
import { d3_customEvent, d3_dispatch, d3_event, d3_select } from '../../../core/d3-constants';
import { d3_mouse } from '../../../core/d3/mouse';
import { mathMin, mathMax } from '../../../core/zc-constants';
import { $Browser } from '../../../core/browser-detect';

/**
 *
 * @param target
 * @param type
 * @param selection
 * @constructor
 */
function BrushEvent(target, type, selection) {
    this.target = target;
    this.type = type;
    this.selection = selection;
}

function nopropagation() {
    d3_event().stopImmediatePropagation();
}

function preventDefault() {
    let event = d3_event();
    if (event.cancelable) {
        event.preventDefault();
    }
}

/**
 *
 * @param event
 * @return {Touch[] | TouchList}
 */
function isTouchEvent(event) {
    return event.touches;
}

//***********************************/

function d3_eventSource() {
    let e = d3_event(),
        s;
    while ((s = e.sourceEvent)) {
        e = s;
    }
    return e;
}

function defaultFilter() {
    let event = d3_event();
    return !event.ctrlKey && !event.button;
}

/**
 *
 * @param container
 * @param i
 * @return {[number, number]|number[]}
 */
function d3_mousePoint(container, i) {
    let e = d3_eventSource();
    if (e.changedTouches) {
        e = e.touches[i];
    }
    let svg = container.ownerSVGElement || container;
    if (svg.createSVGPoint) {
        let point = svg.createSVGPoint();
        point.x = e.clientX;
        point.y = e.clientY;
        point = point.matrixTransform(container.getScreenCTM().inverse());

        let mouse = [point.x, point.y];

        if (isTouchEvent(d3_event()) && !$Browser.ANDROID && $Browser.uiwebview) {
            //d3.mouse(node) considers pageXOffset and pageYOffset in android devices
            mouse[0] += window.pageXOffset;
            mouse[1] += window.pageYOffset;
        }
        return mouse;
    }
    let rect = container.getBoundingClientRect();
    return [e.clientX - rect.left - container.clientLeft, e.clientY - rect.top - container.clientTop];
}

/**
 *
 * @param brushType
 * @return {brush}
 */
function brushWrapper(brushType) {
    let extent,
        listeners = d3_dispatch('start', 'brush', 'end'),
        filter = defaultFilter;

    /**
     *
     * @param selection
     * @param type
     * @param group
     */
    function emit(selection, type, group) {
        let brushEvent = new BrushEvent(brush, type, selection);
        d3_customEvent(brushEvent, listeners.apply, listeners, [type, group, [brushEvent, group]]);
    }

    function brush(group) {
        let rect = createElement('rect', '.selection', [1], group).attrs({
            fill: '#777',
            'fill-opacity': 0.3,
            stroke: '#fff',
            'shape-rendering': 'crispEdges'
        });

        /**
         *
         * @param p1
         * @param p2
         * @return {[[number, number], [number, number]]}
         */
        const getSelection = (p1, p2) => {
            let x1 = mathMin(p1[0], p2[0]),
                y1 = mathMin(p1[1], p2[1]),
                x2 = mathMax(p1[0], p2[0]),
                y2 = mathMax(p1[1], p2[1]);

            x1 = mathMax(x1, extent[0][0]);
            y1 = mathMax(y1, extent[0][1]);
            x2 = mathMin(x2, extent[1][0]);
            y2 = mathMin(y2, extent[1][1]);

            return [
                [x1, y1],
                [x2, y2]
            ];
        };

        /**
         *
         * @param selection
         * @return {*[]|*}
         */
        const getUserSelection = selection => {
            switch (brushType) {
                case 'X': {
                    return [selection[0][0], selection[1][0]];
                }
                case 'Y': {
                    return [selection[0][1], selection[1][1]];
                }
                default: {
                    return selection;
                }
            }
        };

        /**
         *
         * @param selection
         */
        const drawPreview = selection => {
            let x = null,
                y = null,
                width = null,
                height = null;

            if (selection !== null) {
                switch (brushType) {
                    case 'X': {
                        x = selection[0][0];
                        width = selection[1][0] - selection[0][0];
                        y = extent[0][1];
                        height = extent[1][1] - extent[0][1];
                        break;
                    }
                    case 'Y': {
                        x = extent[0][0];
                        width = extent[1][0] - extent[0][0];
                        y = selection[0][1];
                        height = selection[1][1] - selection[0][1];
                        break;
                    }
                    default: {
                        x = selection[0][0];
                        y = selection[0][1];
                        width = selection[1][0] - selection[0][0];
                        height = selection[1][1] - selection[0][1];
                    }
                }
            }

            rect.attrs({ width, height, x, y });
        };

        const updateSelectionInternal = selection => {
            if (selection !== null) {
                switch (brushType) {
                    case 'X': {
                        selectionInternal[0][0] = selection[0];
                        selectionInternal[1][0] = selection[1];
                        break;
                    }
                    case 'Y': {
                        selectionInternal[0][1] = selection[0];
                        selectionInternal[1][1] = selection[1];
                        break;
                    }
                    default: {
                        selectionInternal = [...selection];
                    }
                }
            }
        };

        let startPoint,
            endPoint,
            windowSelect = d3_select(window),
            selectionInternal,
            eventNode = group.node();

        /**
         *
         */
        const move = function() {
            preventDefault();
            nopropagation();

            let currentPoint;
            if (isTouchEvent(d3_event())) {
                startPoint = d3_mousePoint(eventNode, 0);
                currentPoint = d3_mousePoint(eventNode, 1);
            } else {
                currentPoint = d3_mouse(eventNode);
            }

            // for calling the end event
            endPoint = currentPoint;
            selectionInternal = getSelection(startPoint, currentPoint);
            drawPreview(selectionInternal);

            // fire the event
            emit(getUserSelection(selectionInternal), 'brush', group);
        };

        /**
         *
         */
        const ended = function() {
            preventDefault();
            // we should remove the event with exact name which was used to bind it.
            windowSelect.on('mousemove.brush', null);
            windowSelect.on('touchmove.brush', null);
            windowSelect.on('mouseup.brush touchend.brush', null);
            if (selectionInternal) {
                emit(getUserSelection(selectionInternal), 'end', group);
            }
        };

        group.on('mousedown.brush, touchstart.brush', function() {
            if (!filter.apply(this, arguments)) {
                return;
            }

            preventDefault();
            // touch zoom
            // nopropagation();

            startPoint = d3_mouse(eventNode);
            selectionInternal = [startPoint, startPoint];
            drawPreview(selectionInternal);
            emit(getUserSelection(selectionInternal), 'start', group);
            //
            let event = d3_event(),
                touches = event.touches;
            // need to capture the event while
            if (isTouchEvent(event) && touches.length > 1) {
                windowSelect.on('touchmove.brush', move, true);
            } else {
                windowSelect.on('mousemove.brush', move, true);
            }
            // bind the event
            windowSelect.on('mouseup.brush touchend.brush', ended, true);
        });

        brush.move = function(group, selection) {
            if (selection) {
                emit(getUserSelection(selectionInternal), 'start', group);

                // update with the current brush
                updateSelectionInternal(selection);
                emit(getUserSelection(selectionInternal), 'brush', group);
                emit(getUserSelection(selectionInternal), 'end', group);
            } else {
                selectionInternal = null;
            }
            drawPreview(selectionInternal);
        };
    }

    brush.extent = function(_) {
        return arguments.length ? (extent = _) : extent, brush;
    };

    brush.on = function() {
        let value = listeners.on.apply(listeners, arguments);
        return value === listeners ? brush : value;
    };

    // error handle
    brush.handleSize = () => {};

    // #ZC2380, #ZC2380
    brush.filter = function(_) {
        return arguments.length ? ((filter = typeof _ === 'function' ? _ : filter), brush) : filter;
    };

    return brush;
}

export function brushX() {
    return brushWrapper('X');
}

export function brushY() {
    return brushWrapper('Y');
}

export function brush() {
    return brushWrapper('XY');
}
