// @zohocharts/zc/src/boxplot/boxplot-highlight.js
import { BarHighlighter } from '../bar/bar-highlight';
import { inherit, isArray, _super } from '../utils/zcutils';
import { getCorrectedRectParams } from '../utils/zcsvgutils';
import { RECT_ELEMENT } from '../core/element-constants';
import { patternFillEffect } from '../events/visualEffects/highlight/pattern-fill';
import { HIGHLIGHT_CONTEXT } from '../events/visualEffects/highlight/highlight-context';
import { renderDom } from '../mark/dom-render';
import { ParserRegistry } from '../core/parser-registry';

export function BoxplotHighlighter(chart, renderer) {
    this.chart = chart;
    this.renderer = renderer;
}

inherit(BoxplotHighlighter, BarHighlighter);

var proto = BoxplotHighlighter.prototype;

proto._defaultH = function(point) {
    return [
        {
            shape: 'whisker',
            data: proto.data,
            segments: proto.segments,
            path: this.whiskerPath,
            context: this.renderingContext,
            translate: this.translateContext
        },
        {
            shape: 'median',
            data: proto.data,
            segments: proto.segments,
            path: this.medianPath,
            context: this.renderingContext,
            translate: this.translateContext
        },
        {
            shape: RECT_ELEMENT,
            path: _super(this).path,
            data: _super(this).data,
            context: _super(this).renderingContext,
            translate: _super(this).translateContext,
            styles: patternFillEffect
        }
    ];
};

proto.whiskerPath = function(point, model) {
    var whisker = ParserRegistry.marks('boxplot.whisker').call(this.renderer.commonRendererProp().parser, {data: [point]}, "highlight",true);
    renderDom(whisker, model.context.call(this));
};
proto.medianPath = function(point, model){
    var median = ParserRegistry.marks('boxplot.median').call(this.renderer.commonRendererProp().parser,  {data: [point]}, "highlight",true);
    renderDom(median, model.context.call(this));
}
proto.renderingContext = function() {
    var renderingContext = HIGHLIGHT_CONTEXT.get('scatter').call(this, this.chart);
    renderingContext.node().setAttribute("transform","translate(0,0)");
    return renderingContext;
};
proto.box = function(point) {
    var chart = this.chart,
        renderer = this.renderer,
        rp = this.renderer.commonRendererProp().params,
        dataset = this.chart.dataset,
        x = dataset.getX(point),
        y = dataset.getY(point, point.yaxisOrder),
        min = isArray(y) ? y[0] : rp.yscale.domain()[0],
        max = isArray(y) ? y[1] : y,
        whiskers = rp.whiskers,
        accessor = chart.dataObject.getTransRectValue,
        option = renderer.commonOption,
        whisker0,
        whisker1;

    if (whiskers.enabled) {
        whisker0 = point[whiskers.dataindex][0];
        whisker1 = point[whiskers.dataindex][1];
        min = min < whisker0 ? min : whisker0;
        max = max > whisker1 ? max : whisker1;
    }

    if (!chart.cache.data[point._key]) {
        chart.cache.data[point._key] = {};
    }
    var cache = chart.cache.data[point._key];
    cache._box = point._box =
        cache._box ||
        (whiskers.enabled
            ? getRectForlevels(chart, option, rp, constructPoint(point, [min, max], dataset), point)
            : getRectForlevels(chart, option, rp, point, point));
    return point._box;
};
proto.levels = function(point, needAllLevels) {
    //needAllLevels - for automation purpose
    var chart = this.chart,
        renderer = this.renderer,
        rp = this.renderer.commonRendererProp().params,
        median = rp.median,
        whiskers = rp.whiskers,
        dataset = this.chart.dataset,
        x = dataset.getX(point),
        y = dataset.getY(point, point.yaxisOrder),
        medianVal = point[median.dataindex],
        whisker0 = whiskers.enabled && point[whiskers.dataindex][0],
        whisker1 = whiskers.enabled && point[whiskers.dataindex][1],
        medianT = rp.yscale(medianVal),
        t = whiskers.strokeWidth * 5, //medianThickness
        k = rp.axisrotated ? -1 : 1, //multiply factor
        accessor = chart.dataObject.getTransRectValue,
        option = renderer.commonOption,
        p1 = constructPoint,
        inv = rp.yscale.invert; //To find the differentiate of median, we are using invert function; Need to check for ordinal datatype

    if (!chart.cache.data[point._key]) {
        chart.cache.data[point._key] = {};
    }
    var cache = chart.cache.data[point._key];
    if (!cache._levels) {
        if (!whiskers.enabled && !median.enabled) {
            cache._levels = point._levels = null;
            return point._levels;
        }
        let w0,
            w1,
            q0,
            q1,
            m,
            _levels = [];

        q0 = getRectForlevels(chart, option, rp, p1(point, [y[0], inv(medianT + t * k)], dataset), point, 'high');
        q1 = getRectForlevels(chart, option, rp, p1(point, [inv(medianT - t * k), y[1]], dataset), point, 'low');

        _levels.push(q0, q1);

        if (whiskers.enabled || needAllLevels) {
            w0 = getRectForlevels(chart, option, rp, p1(point, [whisker0, y[0]], dataset), point, 'whisker0');
            w1 = getRectForlevels(chart, option, rp, p1(point, [y[1], whisker1], dataset), point, 'whisker1');

            _levels.push(w0, w1);
        }
        if (median.enabled || needAllLevels) {
            m = getRectForlevels(chart, option, rp, p1(point, [inv(medianT + t * k), inv(medianT - t * k)], dataset), point, 'median');

            _levels.push(m);
        }

        cache._levels = point._levels = _levels;
    }

    return point._levels;
};

function getRectForlevels(chart, option, rp, point, point0, type) {
    let x = chart.dataObject.getTransRectValue('x', point, 0, 0, option),
        y = chart.dataObject.getTransRectValue('y', point, 0, 0, option),
        w = chart.dataObject.getTransRectValue('width', point, 0, 0, option),
        h = chart.dataObject.getTransRectValue('height', point, 0, 0, option),
        transPoint = getCorrectedRectParams(x, y, w, h, null, rp.yaxisreversed, rp.axisrotated);

    return {
        _data: function() {
            return {
                _x: transPoint.x1,
                _y: transPoint.y1,
                _width: w,
                _height: h,
                _widthT: transPoint.x2,
                _heightT: transPoint.y2,
                _point: point0,
                type: type,
                _uid: point._uid
            };
        }
    };
}

function constructPoint(point, newY, dataset) {
    var p1 = [],
        xidx = dataset.xidx,
        yidx = dataset.yidx(point.yaxisOrder),
        x = dataset.getX(point);
    (p1[xidx] = x), (p1[yidx] = newY);
    p1.index = point.index;
    return p1;
}

BarHighlighter.prototype.box = proto.box;
BarHighlighter.prototype.levels = proto.levels;
