// @zohocharts/zc/src/boxplot/boxplot-whisker-encode.js
import { getBarMarkerSize } from '../bar/bar-marker-helpers';
import { defined, pick } from '../utils/zcutils';
import { whiskerSize } from './whiskers';

export let boxplot_whisker_position = function() {
    let position_helpers = {};

    const KEY = 'whiskers',
        KEY_MERGED = 'whiskersOptionsMerged';

    if (this.builderC[KEY] && this.builderC[KEY].enabled) {
        let builderC = this.builderC,
            whiskerOptions = builderC[KEY_MERGED],
            yidx = this.chart.dataset.yidx(this.commonOption.yaxiscolumnorder),
            whiskerIdx = whiskerOptions.dataindex,
            isWhiskerOnEdges = builderC.whiskers.showOn == 'edges',
            barMSize = getBarMarkerSize.call(this, KEY_MERGED);

        this.helpers.position[KEY] = {
            start: (d, encode) => {},
            end: (d, encode, order) => {
                let x = this.scales.x(this.chart.dataset.getX(d)) + this.commonOption.levelMarkerWidth / 2;
                x = x + this.scales.interX(this.commonOption.barDatasetIndex) + this.commonOption.padding;

                let y = defined(order)
                    ? this.scales.y(d[whiskerIdx][order])
                    : (this.scales.y(d[whiskerIdx][0]) + this.scales.y(d[whiskerIdx][1])) / 2;
                if (encode == 'size') {
                    let wSize, customMarkers;
                    if (isWhiskerOnEdges) {
                        wSize = whiskerSize(d, this.builderC.yscale, whiskerIdx, yidx, order);
                        wSize = this.chart.axes.rotated ? -wSize : wSize;
                    } else {
                        wSize = whiskerSize(d, this.builderC.yscale, null, null, null, whiskerIdx);
                    }
                    customMarkers = {
                        symbol: 'vertical-line-z',
                        size: wSize
                    };

                    return [barMSize, wSize, customMarkers];
                }
                return {
                    x,
                    y,
                    rotate: this.chart.rotated ? 90 : 0
                };
            }
        };
    }

    return position_helpers;
};

export function boxplot_whisker_styles() {
    const KEY = 'whiskers',
        KEY_MERGED = 'whiskersOptionsMerged';

    let options = this.builderC[KEY_MERGED];

    if (this.builderC[KEY] && this.builderC[KEY].enabled) {
        this.helpers.styles[KEY] = {
            fill: options.color,
            stroke: options.color,
            strokeWidth: pick(options.strokeWidth, 1),
            strokeDasharray: options.dashStyle,
            strokeLinecap: options.lineCap,
            display: (d, i) => {
                return '';
            }
        };
    }
}
