// @zohocharts/zc/src/boxplot/boxplotOptions.js
import { defaultTheme as options } from '../core/defaulttheme';

options.boxplot = function() {
    return {
        outerPadding: 0.04,
        fillOpacity: 0.9,
        maxBandWidth: 200,
        multiColoring: false,
        border: {
            show: false,
            size: null,
            style: 'solid',
            radius: 0,
            color: null
        },
        whiskers: {
            enabled: true,
            columnindex: 3,
            dataindex: 3,
            size: '50%',
            color: '#d8d8d8',
            opacity: 0.9,
            strokeWidth: 1,
            showOn: 'edges', //'overlay'
            events: {
                enabled: true
            },
            hoveroptions: {
                strokeWidth: 2,
                opacity: 0.9,
                color: '#d8d8d8',
                size: '60%'
            }
        },
        median: {
            enabled: true,
            dataindex: 2,
            columnindex: 2,
            color: '#d8d8d8',
            opacity: 0.9,
            strokeWidth: 1,
            hoveroptions: {
                strokeWidth: 2,
                opacity: 0.9,
                color: '#d8d8d8',
                size: '60%'
            }
        },
        datalabels: {
            showAs: 'y' //x,y,percent
        },
        animation: {
            type: 'verticalAll' //verticalAll || vertical
        },
        gradients: {
            //linear || none
            options: {
                linear: {
                    y2: 50
                }
            }
        }
    };
};

export var _boxplotOptions = options.boxplot;
