// @zohocharts/zc/src/boxplot/boxplot-median-encode.js
import { getBarMarkerSize } from '../bar/bar-marker-helpers';
import { pick } from '../utils/zcutils';

export let boxplot_median_position = function() {
    let position_helpers = {};

    const KEY = 'median',
        KEY_MERGED = 'medianOptionsMerged';

    let builderC = this.builderC,
        medianOptions = builderC[KEY_MERGED],
        medianIdx = medianOptions.dataindex,
        barMSize = getBarMarkerSize.call(this, 'medianOptionsMerged');

    if (this.builderC[KEY] && this.builderC[KEY].enabled) {
        this.helpers.position[KEY] = {
            start: (d, encode) => {},
            end: (d, encode) => {
                let x = this.scales.x(this.chart.dataset.getX(d)) + this.commonOption.levelMarkerWidth / 2;
                x = x + this.scales.interX(this.commonOption.barDatasetIndex) + this.commonOption.padding;

                let y = this.scales.y(d[medianIdx]);

                if (encode == 'size') {
                    return [barMSize, null, null];
                }
                return {
                    x,
                    y,
                    rotate: this.chart.rotated ? 90 : 0
                };
            }
        };
    }

    return position_helpers;
};

export function boxplot_median_styles() {
    const KEY = 'median',
        KEY_MERGED = 'medianOptionsMerged';

    let options = this.builderC[KEY_MERGED];

    if (this.builderC[KEY] && this.builderC[KEY].enabled) {
        this.helpers.styles[KEY] = {
            fill: options.color,
            stroke: options.color,
            strokeWidth: pick(options.strokeWidth, 1),
            strokeDasharray: options.dashStyle,
            strokeLinecap: options.lineCap,
            display: (d, i) => {
                return '';
            }
        };
    }
}
