// @zohocharts/zc/src/packedbubble/packedbubblechart.js
//$Id$
//packedbubblechart.js
import { Registry } from '../core/zc-registry';
import { svg_plot_renderer, labelRenderer, legend_helpers as L_H } from '../core/main-zc-init';
import { d3_pack, d3_select, d3_ascending, d3_ease_linear, d3_event, d3_hierarchy } from '../core/d3-constants';
import { getFillColor, isMultiColoring, getRawColor, getStrokeColor } from '../utils/colorutils';
import { getEventConf, getEventObject } from '../events/events-utils';
// import { legend_helpers as L_H } from '../legend/legend-helper';
import { setPointerEvents } from '../events/events-utils';
import { validateNullVal, defined, isArray, getPropVal } from '../utils/zcutils';
import { getAnimationConf } from '../utils/animationutils';
import { PACKEDBUBBLE_CHARTNAME } from '../core/chartname-constants';
import { FADE_ANIMATION, FADEALL_ANIMATION } from '../core/animation-constants';
import { appendEle } from '../utils/zcsvgutils';
import { CIRCLE_ELEMENT, PATH_ELEMENT } from '../core/element-constants';
import { NONE, TRANSPARENT, NULL, POINTER } from '../core/zc-constants';
import { generateDataForAnnotation } from '../annotation/annotation-utils';
import { PERCENT_SHOWAS } from '../label/showas-constants';
import { mathMin, mathAbs } from '../core/zc-constants';
import { getSVGsymbol } from '../utils/svg-symbols';
import { HIGHLIGHT_EFFECT_REGEX } from '../core/main-zc-init';
import { attachevents } from '../events/eventActions';
import { getSvgTransform } from '../utils/svg-transforms';
import { updateSelectionOnDatapoint } from '../events/event-update-selection';
import { isFacet } from '../axis/hierarchy-axis-utils';
import { setOpacityForTickFilterElements, setPivotAttributes } from '../plot/svg/set-pivot-attr';

let annotation = Registry.getComponent('annotation');
/**
 *
 * @returns {chart} packedbubble renderer.
 */
svg_plot_renderer.packedbubble = function() {
    var pack = d3_pack(),
        isredraw = false,
        commonRendererProp,
        bubbleGroup,
        path,
        chartObj,
        commonPlotOptions,
        mode,
        seriesGroup = [],
        zchighlighterPosition,
        highlightGroup,
        highlightEle,
        multiColoring,
        valueFunction,
        includeDisabledSeries;

    function chart(selection) {
        selection.each(function(data) {
            //renderer config values
            var rp = commonRendererProp.params;
            chartObj = rp.chartObj;
            var color = rp.color,
                notesArray = rp.notesArray,
                yaxiscolumnorder = rp.yaxiscolumnorder,
                dataset = rp.dataset,
                systemconf = chartObj.systemConf,
                labelQueue = rp.labelQueue,
                { bound, rendererIndex = 0 } = rp;

            commonPlotOptions = rp.commonPlotOptions;
            var paneid = isFacet(chartObj) && `${data[0].row}_${data[0].column}`,
                fillColor = [],
                strokeColor = [];

            var cloneddata = data.filter(function(d, i) {
                    let rendererConf = chartObj.rendererConf[rendererIndex + i];
                    d.seriesIndex = i;
                    seriesGroup[i] = [];
                    fillColor[i] = chart.fillColorFunc = getFillColor(chartObj, i, paneid, null, null, null, null, rendererConf.params);
                    strokeColor[i] = getStrokeColor(chartObj, i, paneid, null, rendererConf.params);
                    return !d.disabled;
                }),
                chartWidth = bound.width,
                chartHeight = bound.height,
                isSingleSeries = cloneddata.length === 1,
                isPatternPalleteEnabled = L_H.hasPatternPallete(chartObj) || chartObj.patternPallete,
                notesOptions = chartObj.systemConf.notes,
                notesEnabled = notesOptions && notesOptions.enabled === true && notesOptions.plotValues;
            mode = commonPlotOptions.mode;
            multiColoring = isMultiColoring(commonPlotOptions, chartObj);

            var NegativeAsAbsolute = commonPlotOptions.showAsNegative === 'absolute',
                interPadding = +commonPlotOptions.interPadding || 0,
                outerPadding = +commonPlotOptions.outerPadding || 0,
                diameter = mathMin(chartWidth, chartHeight) - outerPadding;

            zchighlighterPosition = [
                chartObj.plotarea.left + (chartWidth - diameter) / 2 + rp.bound.x,
                chartObj.plotarea.top + (chartHeight - diameter) / 2 + rp.bound.y
            ];
            valueFunction = function(d) {
                if (!includeDisabledSeries && isDisabled(data, d)) {
                    return 0;
                }
                var y = dataset.getY(d, yaxiscolumnorder) || 0;
                if (y < 0) {
                    return NegativeAsAbsolute ? mathAbs(y) : 0;
                }
                return y;
            };

            var plotLevelOptions = d => d.depth && data[getSeriesPos(d)].plotoptions,
                isEventEnabled = d => {
                    var plotOptions = plotLevelOptions(d);
                    return validateNullVal(getEventConf(plotOptions, commonPlotOptions, chartObj, 'enabled'), true);
                },
                cursor = d => {
                    var plotOptions = plotLevelOptions(d);
                    return isEventEnabled(d) ? getEventConf(plotOptions, commonPlotOptions, chartObj, 'cursor') || POINTER : NULL;
                };

            //Animation variables
            var animationOptions = getAnimationConf(commonPlotOptions, null, chartObj, PACKEDBUBBLE_CHARTNAME);
            var isAnimationEnable = animationOptions.enabled,
                animationType = animationOptions.type,
                easingType = animationOptions.easingType,
                transitionDuration = isredraw ? (isAnimationEnable ? 200 : 0) : animationOptions.duration;

            //Transition variables
            var initial_transform = function(d) {
                    var transform = d3_select(this).attr('transform');
                    return transform != null ? transform : `translate(${diameter / 2},${diameter / 2})`;
                },
                final_transform = function(d) {
                    return `translate(${d.x},${d.y})`;
                },
                initial_r = function(d) {
                    var r = d3_select(this).attr('r');
                    return r != null ? r : 0;
                },
                final_r = function(d) {
                    return d.r;
                },
                final_fillOpacity = function(d, i) {
                    return commonPlotOptions.fillOpacity;
                },
                final_strokeOpacity = function(d, i) {
                    return commonPlotOptions.strokeOpacity;
                },
                isFade = !isredraw && (animationType === FADE_ANIMATION || animationType === FADEALL_ANIMATION);

            //Initializing pack layout
            pack.size([diameter, diameter]).padding(interPadding);

            bubbleGroup = appendEle(selection, 'g', [1], '.', 'class', 'hierarchychart');
            bubbleGroup.attr('transform', `translate(${(chartWidth - diameter) / 2},${(chartHeight - diameter) / 2})`);
            path = appendEle(bubbleGroup, CIRCLE_ELEMENT, getPackNodeDescendants(data, true));

            path.attrs({
                class: d => `zc_packedbubble series_${getSeriesPos(d)}`
            }).style('visibility', function(d) {
                return isDisabled(data, d) ? 'hidden' : null;
            });

            path = path.filter(function(d) {
                return !isDisabled(data, d);
            });
            var packedData = getPackNodeDescendants(cloneddata);
            chart.total = packedData[0].value;
            path = path.data(packedData);

            path.attrs({
                transform: isFade ? final_transform : initial_transform,
                r: isFade ? final_r : initial_r
            }).styles({
                fill: function(d, i) {
                    let seriesIndex = getSeriesPos(d);

                    d.index = i;
                    d.data.paneid = paneid;
                    //Test_654.json
                    d.r = d.r || 0;
                    d.x = d.x || 0;
                    d.y = d.y || 0;

                    if (isFacet(chartObj)) {
                        d.data._parent = d.parent;
                    } else {
                        d.data.parent = d.parent;
                    }

                    if (notesEnabled) {
                        d.data.x = d.x;
                        d.data.y = d.y;
                    }

                    if (!d.children && d.parent) {
                        // Test_655.json
                        seriesGroup[seriesIndex].push(d);
                    }

                    if (!d.depth) {
                        d.arcColor = mode == 'flat' || multiColoring || isSingleSeries ? NONE : 'grey';
                    } else if (mode !== 'flat' && d.depth === 1 && multiColoring) {
                        d.arcColor = 'grey';
                    } else {
                        d.arcColor = isFacet(chartObj)
                            ? getRawColor(chartObj, multiColoring ? getItemPos(d) : seriesIndex, d.data, seriesIndex, paneid)
                            : color[(multiColoring ? getItemPos(d) : seriesIndex) % color.length];
                    }

                    var baseClr = getBaseClr(d);
                    return (d.fillColor =
                        (!d.depth && isPatternPalleteEnabled) || baseClr === NONE
                            ? NONE
                            : fillColor[seriesIndex || 0](multiColoring ? getItemPos(d) : seriesIndex, d.data, null, baseClr));
                },
                'fill-opacity': isFade ? 0 : final_fillOpacity,
                'stroke-opacity': isFade ? 0 : final_strokeOpacity,
                stroke: function(d) {
                    let seriesIndex = getSeriesPos(d);

                    if (d.arcColor === NONE) {
                        return NONE;
                    } else if (commonPlotOptions.strokeColor) {
                        return commonPlotOptions.strokeColor;
                    } else if (d.children) {
                        return d.arcColor;
                    } else {
                        return strokeColor[seriesIndex || 0](multiColoring ? getItemPos(d) : seriesIndex, d.data);
                    }
                },
                'stroke-width': function(d) {
                    return commonPlotOptions.strokeWidth;
                },
                cursor: cursor,
                'pointer-events': d => (isEventEnabled(d) && !d.children ? null : NONE)
            });

            //Animation
            var total = path.size(),
                dur = !isredraw && animationType === FADE_ANIMATION ? transitionDuration / total : transitionDuration,
                del = animationType === FADE_ANIMATION ? dur : 0;

            path.transition()
                .delay(function(d, i) {
                    return isredraw ? 0 : i * del;
                })
                .duration(dur)
                .ease(isredraw ? d3_ease_linear : easingType)
                .attrs({
                    transform: final_transform,
                    r: final_r
                })
                .styles({
                    'fill-opacity': final_fillOpacity,
                    'stroke-opacity': final_strokeOpacity
                });

            setPivotAttributes(path, chartObj, 'packedbubble');
            setOpacityForTickFilterElements(path, chartObj, 'packedbubble');

            //Events
            setPointerEvents(chartObj, path, d => !isEventEnabled(d), paneid);
            if (!chartObj.cache.search) {
                chartObj.cache.search = {};
            }
            var mousemove = function(e, d) {
                    var args = getEventArgs(e, d);
                    chartObj.eventHandler.handlers.mousemove.call(chartObj.eventHandler, e, args);
                    updateSelectionOnDatapoint(d, args);
                    chartObj.cache.search.lastPane = paneid;
                },
                mouseout = function(e, d) {
                    var args = getEventArgs(e, d);
                    args.point = d.data;
                    chartObj.eventHandler.handlers.mouseout.call(chartObj.eventHandler, e, args, getSeriesPos(d));
                    updateSelectionOnDatapoint(d, args);
                    chartObj.cache.search.lastPane = null;
                },
                click = function(e, d) {
                    var args = getEventArgs(e, d);
                    chartObj.eventHandler.handlers.click.call(chartObj.eventHandler, e, args, d, getSeriesPos(d), getItemPos(d), chart);
                    updateSelectionOnDatapoint(d, args);
                    chartObj.cache.search.lastPane = paneid;
                },
                dblclick = function(e, d) {
                    if (!chart.chartRendered) {
                        return;
                    }
                    var dt = !annotation._empty
                        ? generateDataForAnnotation(d.data, data.length > 1 ? data[getSeriesPos(d)].seriesname : null)
                        : d.data;
                    var coordinates = [d.x, d.y];
                    coordinates = [coordinates[0] + (chartWidth - diameter) / 2, coordinates[1] + (chartHeight - diameter) / 2];
                    var args = getEventArgs(e, d);
                    args.coordinate = coordinates;
                    args.basedOn = 'plot';
                    args.data = dt;
                    args.point = d.data;
                    chartObj.eventHandler.handlers.doubleclick.call(chartObj.eventHandler, e, args, getSeriesPos(d), d);
                    updateSelectionOnDatapoint(d, args);
                    chartObj.cache.search.lastPane = paneid;
                },
                eventAttacherOptions = {
                    target: 'plot',
                    touchleave: true,
                    chartObj: chartObj
                };
            var mouseHandlers = {
                    mousemove: mousemove,
                    mouseout: mouseout,
                    click: click,
                    dblclick: dblclick
                },
                touchHandlers = {
                    singletouchmove: mousemove,
                    touchend: mouseout,
                    tap: click,
                    dbltap: dblclick
                };
            attachevents(
                path.filter(function(d) {
                    return !d.children;
                }),
                mouseHandlers,
                touchHandlers,
                eventAttacherOptions
            );

            chartObj.timerObj[chartObj.timerObj.length] = chart.renderedTimer = setTimeout(function() {
                //Annotation preprocessing
                if (!annotation._empty) {
                    data.forEach(function(series, i) {
                        notesArray = annotation.preprocessAnnotations(
                            data,
                            i,
                            function(dt, k, j) {
                                var c = [dt.x, dt.y];
                                return [c[0] + (chartWidth - diameter) / 2, c[1] + (chartHeight - diameter) / 2];
                            },
                            notesArray,
                            chartObj,
                            rp
                        );
                    });
                }

                if (labelRenderer) {
                    var transform = [(chartWidth - diameter) / 2, (chartHeight - diameter) / 2],
                        totalFunc = function(d, showAs) {
                            if (showAs === PERCENT_SHOWAS) {
                                return chart.total;
                            } else {
                                return d.parent.value;
                            }
                        };

                    data.forEach(function(d, i) {
                        let _currentRenderer = chartObj.renderer[isFacet(chartObj) ? rendererIndex + i : 0],
                            _currentRendererIndex = isFacet(chartObj) ? rendererIndex + i : i,
                            paneid = isFacet(chartObj) ? `${d.row}_${d.column}` : i;

                        labelQueue.push({
                            renderer: _currentRenderer,
                            charttype: 'hierarchy',
                            chartObj: chartObj,
                            rendererIndex: _currentRendererIndex,
                            paneid,
                            args: [data[i], i, null, null, totalFunc, seriesGroup[i], transform]
                        });
                    });
                }

                isredraw = true;
                chart.chartRendered = true;
            }, transitionDuration + 17);

            function getEventArgs(e, d) {
                var ev = getEventObject(e),
                    seriesIndex = getSeriesPos(d),
                    subSeriesIndex = 0,
                    itemIndex = getItemPos(d);
                return {
                    renderer: chart,
                    seriesIndex: seriesIndex,
                    subSeriesIndex: subSeriesIndex,
                    itemIndex: itemIndex,
                    paneid: paneid,
                    rendererIndex: rendererIndex,
                    coordinates: [ev.pageX, ev.pageY],
                    element: this,
                    data: d,
                    selected: d.selected,
                    _key: `ZC_${chartObj.id}_${seriesIndex}_${subSeriesIndex}_${itemIndex}`
                };
            }
        });
        return chart;
    }

    chart.commonRendererProp = function(_) {
        if (!arguments.length) {
            return commonRendererProp;
        }
        commonRendererProp = _;
        return chart;
    };

    chart.defaultMouseOverFunc = function(d, seriesPos, itemPos, paneid) {
        createHighlightElement();
        var refElement = d3_select(path.nodes()[d.index]);
        var refElementPosition = getSvgTransform(refElement).translate;
        var highlightEffect = chartObj.legendHighlight.getLegendHighlightEffect(seriesPos, 'selectedSeries', 'plot', paneid);
        var currentColor = getRawColor(chartObj, multiColoring ? itemPos : seriesPos, d.data, seriesPos, paneid);

        highlightGroup.style('opacity', null);
        highlightEle
            .attrs({
                transform: `translate(${zchighlighterPosition[0] + refElementPosition[0]},${zchighlighterPosition[1] +
                    refElementPosition[1]})`,
                d: getSVGsymbol('circle')(+refElement.attr('r') * 2)
            })
            .styles(
                chartObj.legendHighlight.getEffect(
                    highlightEffect,
                    PACKEDBUBBLE_CHARTNAME,
                    HIGHLIGHT_EFFECT_REGEX.test(highlightEffect) ? currentColor : TRANSPARENT,
                    { fill: TRANSPARENT, stroke: currentColor },
                    multiColoring ? itemPos : seriesPos,
                    'plot'
                )
            );
    };

    chart.defaultMouseOutFunc = function(d, i) {
        createHighlightElement(); //Touch device -> tap on a data, then tap a legend item. Mouseout should be called and legend tap action also will get executed.

        highlightGroup.style('opacity', 0);
    };

    function createHighlightElement() {
        if (highlightGroup) {
            return;
        }

        var g = chartObj.staticPaneSVG.selectAll('g#zchierarchyhighlight');
        if (!g.empty()) {
            //#ZC1756
            highlightGroup = g;
            highlightEle = appendEle(highlightGroup, PATH_ELEMENT, [1]);
        }
    }

    function getNodes(root) {
        var leafNodes = [];

        function recurse(node, prevID) {
            var child = getChildren(node);
            if (child) {
                child.forEach(function(c, i) {
                    var ID = defined(prevID) ? `${prevID}_${i}` : `${c.seriesIndex}`;
                    parseID(c, ID);
                    recurse(c, ID);
                });
            } else {
                parseID(node, prevID);
                if (mode == 'flat') {
                    leafNodes.push(node);
                }
            }
        }

        function parseID(node, ID) {
            var indices = ID.split(/_/);
            node.seriesPos = +indices[0];
            node.itemPos = +indices[indices.length - 1];
        }

        recurse(root);
        return mode == 'flat' ? leafNodes : root;
    }

    function getChildren(node) {
        var depth = node.level && node.level.split('_').length;
        if (!chartObj.dataObject.hasHierarchyData && depth == 1 && node.length < 2 && node[0] && isArray(node[0][0])) {
            return node[0];
        } else if (isArray(node[0])) {
            if (isArray(node[0][0])) {
                return node;
            } else {
                return node.filter(function(d) {
                    return isArray(d[0]) || valueFunction(d) > 0; //Test_761.json
                });
            }
        } else {
            return null;
        }
    }

    function getSeriesPos(d) {
        var level = d.level || (d.data && d.data.level);
        if (level) {
            return +level.split('_')[0];
        }
    }

    function getItemPos(d) {
        return d.data.itemPos;
    }

    function getBaseClr(d, rendererColor) {
        return d.children ? d.arcColor : rendererColor;
    }

    function isDisabled(seriesData, d) {
        var seriesIndex = getSeriesPos(d);
        return defined(seriesIndex) && seriesData[seriesIndex].disabled;
    }

    function getPackNodeDescendants(data, include) {
        includeDisabledSeries = include;
        var order = commonPlotOptions.order || d3_ascending;
        if (typeof order === 'string') {
            order = getPropVal(window, order);
        }

        var root = d3_hierarchy(getNodes(data), getChildren)
            .sum(valueFunction)
            .sort(order);

        pack(root);
        return root.descendants();
    }

    return chart;
};

export var _packedbubbleChart = svg_plot_renderer.packedbubble;
