// @zohocharts/zc/src/polaraxis/parser/marks/polar-domain-marks.js
//polaraxis domain marks
import { d3_arc } from "../../../core/d3-constants";
import { ParserRegistry } from "../../../core/parser-registry";
import { NONE } from "../../../core/zc-constants";
import { getDashArray, getLineCap } from "../../../utils/dashstyles";
import { validateNullVal } from "../../../utils/zcutils";

export function _polarDomain(data) {
    const self = this;

    const axisline = data.axisline,
        range = this.range,
        renderer = this.renderer,
        axislineSize = axisline.show ? validateNullVal(axisline.strokeWidth, 1) : 0,
        radius = renderer.radius();

    // add default marks here
    //TODO: Maybe animate later
    // const enterMarks = {
    //     d: function () {
    //         return d3ArcPath(radius - 1, radius, range[0], range[0]);
    //     }
    // }, updateMarks = {
    //     d: function () {
    //         return d3ArcPath(radius - 1, radius, range[0], range[1]);
    //     }
    // };


    const domainMarks = {
        mark: 'path',
        selector: '.domain',
        data: axisline.show ? [1] : [],
        anime: {
            // enter: enterMarks,
            // update: updateMarks
        },
        config: {
            strokeWidth: axislineSize,
            d: function () {
                return d3ArcPath(radius - 1, radius, range[0], range[1]);
            }
        },
        properties: {
            styles: {
                stroke: axisline.color,
                fill: NONE,
                'stroke-dasharray': getDashArray(axisline.style, axislineSize),
                'stroke-linecap': getLineCap(axisline.lineCap, axisline.style),
                'pointer-events': NONE
            }
        }
    }
    return domainMarks;
}
ParserRegistry.axis('polar.domain', _polarDomain);

function d3ArcPath(innerRadius, outerRadius, startAngle, endAngle) {
    let path = d3_arc()
        .innerRadius(innerRadius)
        .outerRadius(outerRadius)
        .startAngle(startAngle)
        .endAngle(endAngle)(1);
    return `M${path.split(/L|M/gi)[1]}`;
}
