// @zohocharts/zc/src/polaraxis/parser/marks/polar-ticks-marks.js
import { INNER_ALIGN } from "../../../core/align-constants";
import { ParserRegistry } from "../../../core/parser-registry";
import { NONE, NULL, POINTER, mathPI } from "../../../core/zc-constants";
import { $ZCG } from "../../../core/zcglobal";
import { getDashArray, getLineCap } from "../../../utils/dashstyles";
import { translateString } from "../../../utils/svg-transforms";
import { applyFont } from "../../../utils/zcsvgutils";
import { validateNullVal } from "../../../utils/zcutils";
import { d0_360, getDY, polarToXY } from "../../polarutils";
import { _polarAxisEvents } from "../config/polaraxis-events";

export function _polarTicksGroup(data) {
    const self = this;

    const {chart, ticks, renderer} = this,
        systemconf = chart.systemConf,
        radius = renderer.radius(),
        tickFormat = renderer.tickFormat(),
        scale = renderer.scale(),
        { tickOrient, ticklabel, axisline, tickmark, subtickmark, grid } = data,
        axislineSize = axisline.show ? validateNullVal(axisline.strokeWidth, 1) : 0,
        ticklabelPad = validateNullVal(ticklabel.marginLeft, 5),
        tickmarkSize = validateNullVal(tickmark.size, 3),
        gridSize = validateNullVal(grid.strokeWidth, 1),
        // subtickmarkSize = validateNullVal(subtickmark.size, 2),
        events = ticklabel.events || {},
        eventDisabled = validateNullVal(events.enabled, true) == false,
        cursor = events?.cursor;

    
    let {tickRad, textRad, tickmarkRad, subtickmarkRad} = this.commonOption;

    // add ticklabel events here
    const tickLabelEvents = _polarAxisEvents.call(this, data);

    const ticksMark = {
        mark: 'g',
        selector: '.ticksGroup',
        data: [1],
        properties: {
        },
        children: [{
            mark: 'g',
            selector: '.tick',
            data: ticks,
            properties: {
                attrs: {
                    class: (d, i) => `tick tick_${i}`,
                    transform: function (d) {
                        const xy = polarToXY(tickRad, scale(d));
                        return translateString(xy);
                        // return `translate(${xy[0]},${xy[1]})`;
                    }
                }
            },
            children: [{
                mark: 'line',
                selector: '.tickmark',
                data: d => [d],
                config: {
                    strokeWidth: tickmark.strokeWidth
                },
                properties: {
                    attrs: {
                        x2: function (d) {
                            return polarToXY(tickmarkRad, scale(d))[0];
                        },
                        y2: function (d) {
                            return polarToXY(tickmarkRad, scale(d))[1];
                        }
                    },
                    styles: {
                        stroke: tickmark.color
                    }
                }
            }, {
                mark: 'line',
                selector: '.grid',
                data: d => [d],
                config: {
                    strokeWidth: gridSize
                },
                properties: {
                    attrs: {
                        x2: function (d) {
                            return polarToXY(-radius, scale(d))[0];
                        },
                        y2: function (d) {
                            return polarToXY(-radius, scale(d))[1];
                        }
                    },
                    styles: {
                        stroke: grid.color,
                        'stroke-dasharray': getDashArray(grid.style, gridSize),
                        'stroke-linecap': getLineCap(grid.lineCap, grid.style)
                    }
                }
            }, {
                mark: 'text',
                selector: '.ticklabel',
                data: d => [d],
                config: {
                    text: function (d) { return tickFormat(d) }
                },
                events: { ...tickLabelEvents.mouse, ...tickLabelEvents.touch },
                properties: {
                    attrs: {
                        index: function (d, i) { return i },
                        dy: function (d) {
                            var a = d0_360(scale(d));
                            return tickOrient == INNER_ALIGN
                                ? !(a > mathPI / 2 && a < (3 * mathPI) / 2)
                                    ? `${-getDY(a)}em`
                                    : null
                                : a > mathPI / 2 && a < (3 * mathPI) / 2
                                    ? `${getDY(a)}em`
                                    : null;
                        },
                        'text-anchor': function (d) {
                            var a = d0_360(scale(d));
                            return a % mathPI == 0
                                ? 'middle'
                                : (tickOrient == INNER_ALIGN && a < mathPI) || (tickOrient != INNER_ALIGN && a > mathPI)
                                    ? 'end'
                                    : null;
                        },
                        transform: function (d) {
                            var xy = polarToXY(textRad, scale(d));
                            return translateString(xy);
                        }
                    },
                    styles: {
                        cursor: cursor || POINTER,
                        'pointer-events': eventDisabled ? NONE : NULL
                    }
                }
            }],
            onEnd: function (ele) {
                let text = ele.select('text.ticklabel');
                applyFont(text, [ticklabel, systemconf.chart, systemconf.canvas, $ZCG])
            }
        }]
    }
    return ticksMark;
}
ParserRegistry.axis('polar.ticks', _polarTicksGroup)
