// @zohocharts/zc/src/polaraxis/parser/marks/polar-subticks-marks.js
import { INNER_ALIGN } from "../../../core/align-constants";
import { ParserRegistry } from "../../../core/parser-registry";
import { translateString } from "../../../utils/svg-transforms";
import { validateNullVal } from "../../../utils/zcutils";
import { polarToXY } from "../../polarutils";

export function _polarSubticksGroup(data) {
    const self = this;

    const subticks = this.subticks,
        renderer = this.renderer,
        scale = renderer.scale(),
        { tickOrient, subtickmark } = data,
        subtickColor = subtickmark.color,
        subtickmarkSize = validateNullVal(subtickmark.size, 2);

    let {tickRad, textRad, tickmarkRad, subtickmarkRad} = this.commonOption;

    const subticksMark = {
        mark: 'g',
        selector: '.subticksGroup',
        data: [1],
        properties: {
            styles: {
                stroke: subtickColor
            }
        },
        children: [{
            mark: 'line',
            selector: '.subticks',
            data: subticks,
            properties: {
                attrs: {
                    class: (d, i) => `subticks subticks_${i}`,
                    x2: function (d) {
                        return polarToXY(subtickmarkRad, scale(d))[0];
                    },
                    y2: function (d) {
                        return polarToXY(subtickmarkRad, scale(d))[1];
                    },
                    transform: function (d) {
                        const xy = polarToXY(tickRad, scale(d));
                        return translateString(xy);
                    }
                }
            }
        }]
    }
    return subticksMark;
}
ParserRegistry.axis('polar.subticks', _polarSubticksGroup);
