// @zohocharts/zc/src/legend/expandable/expandable-parser.js
//
import { Registry } from '../../core/zc-registry';
import { POINTER } from '../../core/zc-constants';
import { DIV_ELEMENT } from '../../core/element-constants';
import { EXPANDABLE_CLASSNAME } from '../legend-constants';
import { Mark } from '../../mark/mark-utils';
import { invokeFunction, shorthand_px, setPropVal, defined, pick } from '../../utils/zcutils';
import { applyFont } from '../../utils/zcsvgutils';

function ExpandableParser(spec) {
    let _helper = helper(spec),
        { show, styles, html, expanded, config, onEnd } = _helper;

    let expandableMark = Mark(DIV_ELEMENT, `.${EXPANDABLE_CLASSNAME}`, show ? [1] : []);
    expandableMark.config = { ...config, html: html && html(expanded, null, spec.order) };
    expandableMark._parser = _helper;
    expandableMark._spec = spec; // TODO: check
    expandableMark.onEnd = onEnd;
    setPropVal(expandableMark, 'properties.styles', styles);
    expandableState(expandableMark, expanded);

    return expandableMark;
}

function helper(spec) {
    let {
            config,
            margin = 0,
            bound,
            height,
            padding = 3,
            target,
            remainingCount,
            callbackArgs = [],
            expanded = false,
            isHorizontal,
            priorityArray
        } = spec,
        { show, textAlign, ontoggle, text, keepAligned } = config,
        keepExpandableAligned = keepAligned && isHorizontal,
        html = getContent,
        styles = {
            display: keepExpandableAligned ? 'inline-block' : 'block',
            padding: `${padding}px`,
            // margin: () => shorthand_px(invokeFunction(margin)),
            'text-align': keepExpandableAligned ? null : textAlign, // TODO: Text align for horizontal
            'max-width': `${bound.maxWidth}px`,
            'line-height': keepExpandableAligned && height ? () => `${height()}px` : null,
            cursor: POINTER,
            overflow: 'hidden',
            'text-overflow': 'ellipsis',
            'white-space': 'nowrap',
            ...applyFont(null, [config, ...priorityArray], true)
            // color: config.fontColor,
            // 'font-size': `${utils.getVal(config.fontSize)}px`,
            // 'font-family': config.fontFamily || getValByPriority(priorityArray, 'fontFamily'),
        },
        onEnd = ele => {
            let _target = target(),
                scrollType = isHorizontal ? 'scrollLeft' : 'scrollTop';

            // ele.style('margin', function(){
            //     return shorthand_px(invokeFunction(margin));
            // })
            // On reverse hover scrollTop | scrollLeft will be change. Reset here
            _target.node()[scrollType] = 0;

            if (show) {
                _target.style('overflow', 'hidden');
            } else if (!ele.empty()) {
                ele.remove();
                _target.styles({
                    'overflow-x': 'hidden',
                    'overflow-y': 'hidden'
                });
            }
        }
        remainingCount = spec.remainingCount;

    // show = show && remainingCount > 0;

    if (show && remainingCount > 0) {
        return { show, styles, html, expanded, config, onEnd };
    } else {
        return {
            show: null,
            styles: null,
            html: null,
            expanded: null,
            config: null,
            onEnd: null
        };
    }

    //
    function getContent(expanded, update, order) {
        return function() {
            if (expanded) {
                let _toggleText = invokeFunction(ontoggle, this, expanded, ...callbackArgs, remainingCount, order);

                return _toggleText || 'Show less..';
            } else if (remainingCount > 0) {
                let _toggleText = update ? invokeFunction(ontoggle, this, expanded, ...callbackArgs, remainingCount, order) : null,
                    _text = invokeFunction(text, remainingCount, ...callbackArgs);

                return _toggleText || _text || `${remainingCount} more...`;
            }
        };
    }
}

function expandableState(parser, _) {
    if (defined(_)) {
        setPropVal(parser, 'cache.expanded', _);
    } else {
        return pick(parser && parser.cache && parser.cache.expanded, false);
    }
}

export { ExpandableParser, expandableState };
Registry.setComponent('expandable.parser', ExpandableParser);
