// @zohocharts/zc/src/brush/chart-helpers/domain-helpers.js
import { arrayIncludes, each, firstAndLastItem, isArray, pick, splat } from '../../utils/zcutils';
import { d3_ascending } from '../../core/d3-constants';
import { fixBrushDirection, SelectionToDomain } from '../../behaviours/data-selection/chart/tools/selection-to-domain';
import { scale } from '../../core/main-zc-init';
import { Registry } from '../../core/zc-registry';

export var CHART_DOMAIN_HELPERS = {
    /**
     *
     * @param chart
     * @param range
     * @param type
     * @param index
     * @param chartScale
     * @return {*[]}
     */
    CALCULATE_DOMAIN(chart, range, type, index, chartScale) {
        let datatype = chart.dataset.getDataType(type, index),
            rotated = chart.axes.rotated,
            reversed,
            baseScale;

        if (range === null) {
            if (type === 'x') {
                baseScale = chart.axes.x.baseScale;
            } else if (type === 'y') {
                baseScale = chart.axes.y.groups[index].baseScale;
            }
            return baseScale.domain();
        }

        if (type === 'x') {
            chartScale = pick(chartScale, chart.axes.x.scale);
            reversed = chart.systemConf.chart.axes.xaxis.reversed;
        } else {
            chartScale = pick(chartScale, chart.axes.y.groups[index].scale);
            reversed = chart.systemConf.chart.axes.yaxis[index].reversed;
        }
        var _t = SelectionToDomain(range, chartScale, type, datatype, rotated, reversed, chart, true);
        if (!_t.length) {
            var _d = getEmptyDomain(range, chartScale, type, datatype, rotated, reversed, chart, true);
            chart.emptySelectionDomain = !_d ? chart.emptySelectionDomain : _d;
        }
        return _t;
    },

    /**
     *
     * @param selection
     * @param domainType
     * @param brushType
     * @param index
     * @param chart
     * @return {*[]}
     */
    GET_DOMAIN(selection, domainType, brushType, index = 0, chart) {
        if (brushType === 'xy' && isArray(selection)) {
            if (fixBrushDirection(domainType, chart.axes.rotated) === 'x') {
                selection = [selection[0][0], selection[1][0]];
            } else {
                selection = [selection[0][1], selection[1][1]];
            }
        }
        return this.CALCULATE_DOMAIN(chart, selection, domainType, index);
    },

    /**
     *
     * @param selection
     * @param domainTypes
     * @param brushType
     * @param chart
     * @return {{}}
     */
    GET_DOMAIN_INFO(selection, domainTypes, brushType, chart) {
        let domainInfo = {};

        domainTypes = splat(domainTypes);

        if (arrayIncludes(domainTypes, 'x')) {
            let domain = this.GET_DOMAIN(selection, 'x', brushType, 0, chart);
            domainInfo.x = [{ domain: domain, index: 0, dataType: chart.dataset.getDataType('x', 0) }];
        }

        if (arrayIncludes(domainTypes, 'y')) {
            domainInfo.y = [];
            chart.axes.y.groups.forEach((axis, index) => {
                if (!axis.disabled) {
                    domainInfo.y.push({
                        domain: this.GET_DOMAIN(selection, 'y', brushType, index, chart),
                        index,
                        dataType: chart.dataset.getDataType('y', index)
                    });
                }
            });
        }

        return domainInfo;
    },

    /**
     *
     * @param chart
     * @param domainX
     * @param domainY
     * @return {[]}
     * @constructor
     */
    DOMAIN_TO_DATA: function(chart, domainX, domainY) {
        let xScale = chart.axes.x.scale,
            yScale = chart.axes.y.groups[0].scale,
            xRange,
            yRange,
            points = [];

        if (isArray(domainX)) {
            domainX = firstAndLastItem(domainX);
            xRange = domainX.map(xScale).sort(d3_ascending);
        }

        if (isArray(domainY)) {
            domainY = firstAndLastItem(domainY);
            yRange = domainY.map(yScale).sort(d3_ascending);
        }

        chart.renderer.filter(renderer => {
            let series = renderer.seriesdataAfterUpdate;

            series.forEach(subSeries => {
                let filteredPoints = subSeries.filter(point => {
                    return CHART_DOMAIN_HELPERS.POINT_WITHIN_RANGE(point, xRange, yRange, chart);
                });
                points = points.concat(filteredPoints);
            });
        });
        return points;
    },

    /**
     *
     * @param point
     * @param xRange
     * @param yRange
     * @param chart
     * @param yIndex
     * @return {*}
     * @constructor
     */
    POINT_WITHIN_RANGE: function(point, xRange, yRange, chart, yIndex = 0) {
        let xScale = chart.axes.x.scale,
            yScale = chart.axes.y.groups[yIndex].scale;

        var withinRange = d => {
            let xPass = true,
                yPass = true;

            if (isArray(xRange)) {
                let xValue = xScale(chart.dataset.getX(d));
                xPass = xValue >= xRange[0] && xValue <= xRange[1];
            }

            if (isArray(yRange)) {
                let yValue = yScale(chart.dataset.getY(d, yIndex));
                yPass = yValue >= yRange[0] && yValue <= yRange[1];
            }

            return xPass && yPass;
        };

        return withinRange(point);
    },

    /**
     *
     * @param point
     * @param chart
     * @return {*}
     * @constructor
     */
    POINT_WITHIN_DOMAIN: function(point, chart) {
        let domainInfo = chart.plotBrush.domainInfo,
            xRange;

        let domain = domainInfo.x[0].domain;
        domain = [domain[0], domain[domain.length - 1]];

        if (isArray(domain)) {
            xRange = domain.map(chart.axes.x.scale).sort(d3_ascending);
        }

        return CHART_DOMAIN_HELPERS.POINT_WITHIN_RANGE(point, xRange, null, chart);
    },

    /**
     *
     * @param domains
     * @param chart
     * @constructor
     */
    FILTER_CHART(domains, chart) {
        //Reset cache for data
        chart.cache.data = {};
        chart.cache.search = {};

        if (chart.datalabels.hasFixedPosition) {
            each(domains, (info, type) => {
                chart.cache.correction['bubble-' + type + '-' + (info.index || 0)] = null;
                if (chart.cache.correction.domainCorrection) {
                    chart.cache.correction.domainCorrection['marker-' + type + '-' + (info.index || 0)] = null;
                }
            });
        } else {
            // chart.cache.correction = {};
            for (var attr in chart.cache.correction) {
                if (attr.indexOf('bubble') > -1) {
                    let s = attr.split('-');
                    if (chart.dataset.getDataType(s[1], +s[2]) === 'ordinal') {
                        chart.cache.correction[attr] = null;
                    }
                }
            }
            for (var attr in chart.cache.correction.domainCorrection) {
                if (attr.indexOf('marker') > -1) {
                    let s = attr.split('-');
                    if (chart.dataset.getDataType(s[1], +s[2]) === 'ordinal') {
                        chart.cache.correction.domainCorrection[attr] = null;
                    }
                }
            }
        }

        // chart.cache.correction.domainCorrection = {};
        //reset bullet events
        chart.eventHandler.bulletEvents = null;
        if (chart.updateScale) {
            each(domains, (info, type) => {
                splat(info).forEach((info, i) => {
                    chart.updateScale(null, type, info.domain, info.index, info.range, true);
                });
            });

            chart.plot.renderer.draw(true, false);
            if (!chart.dataObject.flattenPoints) {
                chart.dataObject.flattenData();
            }

            //Need to update the event params. Becoz axes.eventscale is a new scale now
            // chart.eventHandler && chart.eventHandler.updateParams(chart, false);
            chart.drawPlotareaClippath();
            chart.updatePlotareaProp();
            chart.positionEventTrackerElement();

            chart.drawThreshold(true);
        }
    }
};
Registry.setComponent("filterChart", CHART_DOMAIN_HELPERS.FILTER_CHART);

export function getEmptyDomain(selection, chartScale, type = 'x', datatype, rotated = false, reversed = false, chart, isBrush) {
    var [inverted_0, inverted_1] = selection.map((x0, i) => {
            let a = { x: 0, y: 0 };
            a[fixBrushDirection(type, rotated)] = x0;
            return scale.invert(datatype, a, chartScale, rotated, type, reversed, null, { selection: selection }, chart, isBrush);
        }),
        _scale = chart.axes.x.scale,
        _d = _scale.domain();
    if (inverted_0 > inverted_1) {
        [inverted_0, inverted_1] = [inverted_1, inverted_0];
    }
    if (inverted_0 === inverted_1) {
        if (!_d[inverted_0 - 1] || !_d[inverted_0 + 1]) {
            _d[inverted_0 - 1] ? (inverted_1 = null) : (inverted_0 = null);
        }
    }
    if(!_d.length){
        return null;
    }
    return {
        start: _d[inverted_0],
        end: _d[inverted_1]
    };

    return;
}
