// @zohocharts/zc/src/behaviours/visual-effects/chart/controllers/overview.js
import { fixHDPI, getPropVal, invokeFunction, isArray, isObject, isUndefined, mergeJSON, setPropVal, simpleClone } from '../../../../utils/zcutils';
import { createElement } from '../../../../utils/zcsvgutils';
import { mathRound, NONE } from '../../../../core/zc-constants';
import { colorizeImageData, copyCanvasData } from '../../../../utils/canvasutils';
import { CHART_BRUSH_HELPERS } from '../../../../brush/chart-helpers/brush-helpers';
import { charts } from '../../../../core/chart-instance-manager';
import { TIME_DATATYPE } from '../../../../core/datatype-constants';
import { getMergedHandleStyles } from '../../brush-handles/handles';
import { d3_select, d3_set } from '../../../../core/d3-constants';
import { DataProcessor } from '../../../../core/data-processor';
import { getPlotSize } from '../../../../plot/plotSize';
import { svg_clips } from '../../../../effects/svg-clips';
import { overview } from '../../../../core/main-zc-init';

/**
 *
 * @param chart
 */
export function getOverViewUserData(chart) {
    let data = chart.systemConf.chart.scroll.overview.data;

    var overview_plotHeight = chart.systemConf.chart.scroll.overview.plotHeight;
    if(!data){
        data = collectOverviewData(chart, data);
    } else{
        data = isObject(data) ? collectOverviewData(chart, data) : data;
        data = invokeFunction(data, chart, collectOverviewData);
    }
    data.plotHeight = overview_plotHeight;
    return data;
}

function collectOverviewData(chart, data){
    if (!data) {
        data = simpleClone(chart.userdata);

        setPropVal(data, 'chart.marginTop', 0);
        setPropVal(data, 'chart.marginBottom', 0);
        setPropVal(data, 'chart.marginLeft', 0);
        setPropVal(data, 'chart.marginRight', 0);

        // setPropVal(data, 'canvas.intelligence', {});
        setPropVal(data, 'canvas.title.show', false);
        setPropVal(data, 'canvas.subtitle.show', false);

        // disable border and shadow
        setPropVal(data, 'canvas.border.show', false);
        setPropVal(data, 'canvas.shadow.show', false);

        // disable axis
        let yAxes = data.metadata.axes.y || [];
        yAxes.forEach((axis, i) => {
            setPropVal(data, `chart.axes.yaxis[${i}].show`, false);
        });

        // for time auto ticks we will show the axis
        let xAxisConfig = chart.dataset.getColumn('x', 0);
        setPropVal(
            data,
            'chart.axes.xaxis.show',
            xAxisConfig.datatype === TIME_DATATYPE && getPropVal(xAxisConfig, 'time.format.specifier') === 'auto'
        );

        // disable the label
        setPropVal(data, 'chart.axes.xaxis.label.show', false);
    }
    var scrollOrient = chart.scroll.settings.orient,
        axisReversed = data.chart.axes.xaxis.reversed;
    //this check is made inorder to achieve racing - if scrollOrient is loft/right if reversed is true - it will work fine, thus even if reversed not given true in racing data, we set it to true here, also if rversed mentioned in racing data that data itself is taken
    if (scrollOrient === 'right' || scrollOrient === 'left') {
        setPropVal(data, 'chart.axes.xaxis.reversed', !isUndefined(axisReversed) && !chart.racing ? axisReversed : true);
    } else {
        // racing-bottom and top, and other overview cases
        setPropVal(data, 'chart.axes.xaxis.reversed', !isUndefined(axisReversed) && !chart.racing ? axisReversed : false);
    }

    setPropVal(data, '__$overview', true);
    setPropVal(data, 'chart.plot.renderer.mode', chart.systemConf.chart.scroll.overview.renderMode);
    if (scrollOrient) {
        setPropVal(data, 'chart.axes.rotated', scrollOrient === 'left' || scrollOrient === 'right');
    } else {
        setPropVal(data, 'chart.axes.rotated', chart.axes.rotated);
    }

    setPropVal(data, 'chart.plot.datalabels.show', false);
    setPropVal(data, 'chart.plot.events.enabled', false);
    setPropVal(data, 'credits.show', false);
    // #ZC2126
    setPropVal(data, 'notes.enabled', false);
    setPropVal(data, 'chart.scroll.enabled', false);
    // setPropVal(data, 'chart.axes.xaxis.ticklabel.textOverflow', 'none');

    setPropVal(data, 'chart.axes.animation.enabled', false);
    // #ZC2117
    setPropVal(data, 'metadata.axes.label', []);
    // #ZC2106
    setPropVal(data, 'legend.enabled', false);
    setPropVal(data, 'legend.size.enabled', false);

    // get all the chart types
    let types = DataProcessor.helpers.getAllChartTypes(data, 'object');

    // disable the datalabels
    types.forEach(t => {
        var plotoptions = chart.systemConf.chart.plot.plotoptions[t.name];
        if (plotoptions && plotoptions.marker) {
            setPropVal(data, `chart.plot.plotoptions.${t.name}.marker.useSVGMarkers`, false);
        }
        setPropVal(data, `chart.plot.plotoptions.${t.name}.datalabels.show`, false);
        // #ZC2117
        setPropVal(data, `chart.plot.plotoptions.${t.name}.datalabels.stackLabels.show`, false);
        setPropVal(data, `chart.plot.plotoptions.${t.name}.animation.enabled`, false);
        // setPropVal(data, `chart.plot.plotoptions.${t.name}.outerPadding`, 0);
    });
    return data;
}

/**
 *
 * @param overviewChart
 * @param scrollBound
 * @return {CanvasRenderingContext2D}
 */
export function getOverlayCanvas(overviewChart, scrollBound) {
    scrollBound.width = mathRound(scrollBound.width);
    var height = getPlotSize(overviewChart, 'height'),
        width = getPlotSize(overviewChart, 'width');
    let canvas = createElement('canvas', '.overview-grayscale', [1], overviewChart.d3container.select('.canvas-component'), 'svg.static')
        .attrs({
            width: width,
            height: height
        })
        .styles({
            position: 'absolute',
            left: `${overviewChart.plotarea.left}px`,
            top: `${overviewChart.plotarea.top}px`,
            width: `${width}px`,
            height: `${height}px`
        });

    return fixHDPI(canvas.node().getContext('2d'));
}

/**
 *
 * @param chartObj
 * @param scrollDimension
 * @param callback
 */
export async function renderOverviewContext(chartObj, scrollDimension, callback) {
    let userdata = getOverViewUserData(chartObj);

    let overviewContext = chartObj.container.selectAll('div.zcoverview');
    overviewContext.styles({
        position: 'absolute',
        'pointer-events': NONE,
        'box-sizing': 'border-box',
        width: `${scrollDimension.width}px`,
        height: `${scrollDimension.height}px`,
        left: `${scrollDimension.left}px`,
        top: `${scrollDimension.top}px`
    });
    //inorder to manage redraw -the container will be considered as hidden if its display is set to none
    overviewContext.style('display', null);
    chartObj.staticPaneSVG.selectAll('.brush-scroll').style('display', null);

    chartObj.overview = charts(overviewContext.node(), userdata, scrollDimension);

    if (!chartObj.systemConf.chart.scroll.overview.show) {
        overviewContext.style('display', 'none');
        chartObj.staticPaneSVG.selectAll('.brush-scroll').style('display', 'none');
    }
    if (callback) {
        callback(chartObj.overview);
    }
}

/**
 *
 * @param {ChartBrushScroll} scroll
 * @return {number}
 */
export function getHandleMarkerSize(scroll, extent) {
    let considerMarkerSize = d3_set(['lineCircle']),
        resizerSettings = scroll.settings.resizer || {},
        markerType = resizerSettings.type;

    if (resizerSettings.enabled && considerMarkerSize.has(markerType)) {
        let userStyles = resizerSettings[markerType] || {},
            styles = getMergedHandleStyles('x', resizerSettings.type, extent, null, userStyles);

        return styles.size;
    }
    return 0;
}

/**
 *
 * @param chartObj
 * @return {{top: *, left: *, width: plotarea.width, height: plotarea.height}}
 */
export function getScrollBoundRect(chartObj, scroll) {
    let scrollSettings = scroll.settings,
        rotated = chartObj.axes.rotated,
        plotarea = chartObj.plotarea;
    if (scroll.settings.orient) {
        var width = scroll.settings.orient === 'left' || scroll.settings.orient === 'right' ? scrollSettings.height : plotarea.width,
            height = scroll.settings.orient === 'left' || scroll.settings.orient === 'right' ? plotarea.height : scrollSettings.height;
    } else {
        var width = rotated ? scrollSettings.height : plotarea.width,
            height = rotated ? plotarea.height : scrollSettings.height;
    }
    return {
        width,
        height,
        left: scroll.position.x,
        top: scroll.position.y
    };
}

/**
 *
 * @param overviewChart
 */
export function clipOverview(overviewChart) {
    let scroll = this,
        scrollDimension = scroll.dimension,
        scrollData = scroll.settings,
        extent = scroll.brushExtent,
        overviewSettings = scrollData.overview || {},
        axisRotated = overviewChart.axes.rotated,
        defaultSelection = scroll.scrollDomain,
        selectionArgs = {
            scrollType: this.type,
            brushInstance: this.brushInstance,
            scrollScale: this.settings.brushScale,
            chartObj: this.chart
        },
        selection = isArray(scroll.settings.selectionDomain)
            ? scroll.settings.selectionDomain
            : invokeFunction(scroll.settings.selectionDomain, selectionArgs) || defaultSelection,
        renderMode = scrollData.overview.renderMode;
    // update the scale with the overview chart scale
    if (overviewSettings.syncScale !== false) {
        scroll.brushScale = overviewChart.axes.x.scale;
    }

    if (overviewSettings.bindTo === 'plot') {
        extent = getBrushExtent(overviewChart, scrollData, true);
    }

    // scroll.initialSelection = domainToActualRange(overviewChart, selection, 0, 'x', true);

    let brushInstance = scroll.addEvent(extent, selection),
        context,
        canvasData;
    if (renderMode === 'canvas') {
        context = getOverlayCanvas(overviewChart, scrollDimension);
        canvasData = copyCanvasData(overviewChart.context);
        colorizeImageData(canvasData, overviewSettings.unselectedEffect);
    } else {
        this.chart.overview.svg.selectAll('.seriesholder').style('display', 'none');
    }
    const highlightSelected = () => {
        let selection = brushInstance.getSelection(),
            clip = getClipBound(selection, scrollData, axisRotated, overviewChart.plotarea);

        if (renderMode === 'canvas') {
            context.putImageData(canvasData, 0, 0);
            context.clearRect(clip.x, clip.y, clip.width, clip.height);
        } else {
            svg_clips.rect(clip, 'overviewClipPath_0', null, null, null, this.chart, null);
        }
        // clip the track
        CHART_BRUSH_HELPERS.CLIP_TRACK(brushInstance, selection, scroll.chart, 'overview');
    };

    highlightSelected();
    brushInstance.addHook('brush.clipCanvas', highlightSelected);
}
export var duplicateOverview = overview.duplicateOverview = function duplicateOverview(chart) {
    var overview_svg = chart.overview && chart.overview.svg,
        _ele = overview_svg && overview_svg.selectAll('.seriesholder_copy');
    if (overview_svg && !(_ele.node())) {
        var seriesholder = overview_svg.selectAll('.seriesholder'),
            _parent = overview_svg.selectAll('.seriesClippathholder');

        var seriesholder_copy = seriesholder.node().cloneNode(1);
        seriesholder_copy.setAttribute('class', 'seriesholder_copy');

        d3_select(seriesholder_copy)
            .selectAll('path')
            .each(function(d, i) {
                var _e = d3_select(this),
                    unselectedEffect = chart.scroll.settings.overview.unselectedEffect,
                    setStyle = function(ele, attr, attrVal, unselectedEffect, method){
                        if (attrVal != NONE && attrVal != 'transparent') {
                            ele[method](attr, unselectedEffect);
                        }
                    };
                // _e.style("opacity", 1);
                setStyle(_e, 'fill', _e.style('fill'), unselectedEffect, 'style');
                setStyle(_e, 'stroke', _e.style('stroke'), unselectedEffect, 'style');
                setStyle(_e, 'fill', _e.attr('fill'), unselectedEffect, 'attr');
                setStyle(_e, 'stroke', _e.attr('stroke'), unselectedEffect, 'attr');
            });
        _parent.node().insertBefore(seriesholder_copy, seriesholder.node());
        seriesholder_copy = overview_svg.selectAll('.seriesholder_copy');

        seriesholder.style('display', null);
        seriesholder_copy.style('display', null);

        seriesholder.attr('clip-path', 'url(#overviewClipPath_0)');
        var clipPath = chart.scroll.brushInstance._track.attr('clip-path');
        seriesholder_copy.attr('clip-path', clipPath);
    }
}

/**
 *
 * @param chartObj
 * @param scrollData
 * @param withinOverViewPlot
 * @return {*[][]|(number[]|*[])[]}
 */
export function getBrushExtent(chartObj, scrollData, withinOverViewPlot) {
    let plotArea = chartObj.plotarea;
    if (withinOverViewPlot) {
        return [[plotArea.left, plotArea.top], [plotArea.width + plotArea.left, plotArea.height + plotArea.top]];
    } else if(scrollData.orient){
        return [
            [0, 0],
            scrollData.orient === "right" || scrollData.orient === "left" ? [scrollData.height, chartObj.plotarea.height] : [chartObj.plotarea.width, scrollData.height]
        ]
    }
    else {
        return [
            [0, 0],
            chartObj.axes.rotated ? [scrollData.height, chartObj.plotarea.height] : [chartObj.plotarea.width, scrollData.height]
        ];
    }
}

/**
 *
 * @param selection
 * @param scrollData
 * @param axisRotated
 * @param plotArea
 * @return {{x: number, width: number, y: number, height: number}}
 */
function getClipBound(selection, scrollData, axisRotated, plotArea) {
    let selectionDistance = selection[1] - selection[0];
    let x = axisRotated ? 0 : selection[0] - plotArea.left,
        y = axisRotated ? selection[0] - plotArea.top : 0,
        width = axisRotated ? scrollData.height : selectionDistance,
        height = axisRotated ? selectionDistance : scrollData.height;

    return { x, y, width, height };
}
