// @zohocharts/zc/src/behaviours/visual-effects/chart/controllers/plot-reset-button.js
import { CHART_DOMAIN_HELPERS } from '../../../../brush/chart-helpers/domain-helpers';
import { ChartBrushDataFinder } from '../../../data-selection/chart/plot-brush-data-finder';
import { getValByPriority, pick, utils, defined, invokeFunction, mergeJSON, isEmptyObject, isArray, getPropVal } from '../../../../utils/zcutils';
import { attachevents } from '../../../../events/eventActions';
import { CustomButton } from '../../../../button/custom-button';
import { getDashArray, getLineCap } from '../../../../utils/dashstyles';
import { $ZCG } from '../../../../core/zcglobal';
import { svgBoxShadow } from '../../../../utils/zcsvgutils';
import { ClearOldInfo_C } from './clear-old-info';
import { applyEffect } from '../../../../effects/zceffects';
import { d3_event, d3_set } from '../../../../core/d3-constants';

export function PlotResetButtonAction() {
    let eventManager = this,
        chart = eventManager.chart,
        options = chart.systemConf.chart.brush,
        behaviours = eventManager.behaviours,
        plotBrushInstanceMap = behaviours.PLOT_BRUSH,
        plotZoomInstanceMap = behaviours.PLOT_ZOOM,
        domain;

    // reset saved [minRange, maxRange] for both axis
    options.x = options.y = null;

    // #ZC2379 call mouseout before the brush reset
    ClearOldInfo_C.call(eventManager);

    if (plotZoomInstanceMap) {
        plotZoomInstanceMap.instance.reset();
    } else if (chart.scroll && chart.scroll.visible && options.type !== 'y') {
        chart.scroll.visiblePointCount = null;
        chart.scroll.brushInstance.setSelection(chart.scroll.brushInstance.getRange());
    } else {
        domain = ChartBrushDataFinder.apply(eventManager, [plotBrushInstanceMap, null, null, true]);
        plotBrushInstanceMap.instance.brushed = false;
        // filter the to null domain
        CHART_DOMAIN_HELPERS.FILTER_CHART(domain, chart);
    }
    // hide it
    eventManager.resetButton.hide();
    //
    // update the brush extent
    plotBrushInstanceMap && plotBrushInstanceMap.instance.updateExtent([[0, 0], [chart.plotarea.width, chart.plotarea.height]]);
    // #ZC2123
    chart.positionEventTrackerElement();
}

/**
 *
 * @param behaviourEventInstance
 * @constructor
 */
export function ShowBrushResetButton(behaviourEventInstance, point, mouse, event) {
    // show the reset button

    let eventManager = this,
        button = getResetButton(eventManager.chart, eventManager, event);

    if (button) {
        button.show();
    }
}
/**
 *
 * @param {Chart} chart
 * @param {BehaviouralEventManager} eventManager
 * @return {*}
 */

function getResetButton(chart, eventManager, event) {
    let scroll = chart.scroll,
        _chart = chart.systemConf.chart,
        brush = _chart.brush,
        behaviour = _chart.behaviour,
        _scale = chart.axes && chart.axes.x && chart.axes.x.scale,
        _r = chart.overview && chart.overview.axes && chart.overview.axes.x.scale.actualRange ? chart.overview.axes.x.scale.actualRange : _scale && _scale.range(),
        _s = scroll && scroll.brushInstance && scroll.brushInstance.getSelection();

    if(_r && _s){
        var minRange = Number(Math.min(_r[0], _r[1]).toFixed(5)),
        maxRange = Number(Math.max(_r[0], _r[1]).toFixed(5)),
        
        minSelection = Number(Math.min(_s[0], _s[1]).toFixed(5)),
        maxSelection = Number(Math.max(_s[0], _s[1]).toFixed(5));
    }
    

    var commonReset = behaviour && behaviour.resetButton,
        brushReset = brush.resetButton,
        zoomDomain = chart.axes.x && chart.axes.x.zoomScale && chart.axes.x.zoomScale.domain(),
        noOverviewAction = scroll && !(minSelection > minRange || maxSelection < maxRange),
        noZoomAction = zoomDomain && (zoomDomain[0] === 0 && zoomDomain[1] === 100);

    const resetButtonVisible = eventManager.resetButton && eventManager.resetButton.visible;    //ZC4272 - reset y-brush overrides x-scroll also
    const brushed = getPropVal(eventManager.behaviours, "PLOT_BRUSH.instance.brushed");
    const noBrushAction = scroll && scroll.visible && _chart.brush.type === 'y' && !(brushed || resetButtonVisible);

    if (noOverviewAction || noZoomAction || noBrushAction) {
        eventManager.resetButton && eventManager.resetButton.hide();
        return;
    }

    if (!((brush.enabled && brushReset && brushReset.enabled) || (commonReset && commonReset.enabled))) {
        return;
    }

    let resetButtonSettings = mergeJSON(commonReset, brushReset, true);

    let buttonOptions = getButtonOptions(resetButtonSettings, chart, defined(eventManager.behaviours.PLOT_ZOOM)),
        resetButton = (eventManager.resetButton = new CustomButton(
            'custom-button-0',
            buttonOptions.label,
            buttonOptions.button,
            chart.staticPaneSVG.select('.eventgroup')
        ));

    applyEffect(resetButton.group, { config: resetButtonSettings, chartObj: chart });

    // existing class name retain
    resetButton.group.classed('zmbtn', true);

    positionResetButton(chart);
    setResetButtonEvents(eventManager.resetButton, chart, resetButtonSettings, event);
    return resetButton;
}
function setResetButtonEvents(behaviourEvent, chart, resetData, event) {
    var eventsData = resetData.events,
        manager = chart.behaviouralEvents.eventManager;

    behaviourEvent.instance = behaviourEvent.group;
    behaviourEvent.defaultBehaviour = {
        click: (behaviourEvent, event, chartObj) => {
            var eventManager = chartObj.behaviouralEvents.eventManager,
                resetButton = eventManager.resetButton;
            PlotResetButtonAction.call(eventManager, resetButton);
        },
        tap: (behaviourEvent, event, chartObj) => {
            var eventManager = chartObj.behaviouralEvents.eventManager,
                resetButton = eventManager.resetButton;
            PlotResetButtonAction.call(eventManager, resetButton);
        }
    };
    behaviourEvent.customBehaviour = {
        click: eventsData && eventsData.click,
        tap: eventsData && eventsData.tap
    };

    var event = d3_event({...event}),
        listeners = {
            click: () => {
                event.type = 'click';
                _l(chart, event);
            },
            tap: () => {
                event.type = 'tap';
                _l(chart, event);
            }
        };
    manager.resetButton.listenerTypes = d3_set(['click', 'tap']);
    manager.attachListeners(manager.resetButton, listeners);

    var _l = (chart, event) => {
        var manager = chart.behaviouralEvents.eventManager,
            behaviours = manager.getBehaviours(behaviourEvent, event.type),
            allowDefault = true;
        if (isArray(behaviours) && behaviours.length) {
            behaviours.forEach(b => {
                if (b.type === 'custom') {
                    invokeFunction(true, b.behaviour, manager, behaviourEvent, event, chart);
                    allowDefault = event.allowDefault;
                } else if (allowDefault) {
                    b.behaviour.apply(manager, [behaviourEvent, event, chart]);
                }
            });
        }
    };
}

/**
 *
 * @param chart
 */
export function positionResetButton(chart) {
    let _chart = chart.systemConf.chart,
        brush = _chart.brush,
        behaviour = _chart.behaviour,
        commonReset = behaviour && behaviour.resetButton,
        brushReset = brush.resetButton,
        resetButtonSettings = mergeJSON(commonReset, brushReset, true),
        eventManager = chart.behaviouralEvents.eventManager,
        resetButton = eventManager.resetButton;

    if (eventManager.resetButton) {
        let x = pick(
                invokeFunction(resetButtonSettings.x, chart, resetButton.bounds.rect),
                chart.plotarea.left + chart.plotarea.width - resetButton.bounds.rect.width
            ),
            y = pick(invokeFunction(resetButtonSettings.y, chart, resetButton.bounds.rect), 10);

        // hide and translate the button by default
        resetButton.translate([x, y]);
    }
}

/**
 *
 * @param buttonSettings
 * @param chart
 * @param zoomEnabled
 * @return {{button: {'stroke-dasharray': [], 'stroke-width': (*|number), radius: number, fill: *, 'fill-opacity': *, stroke: *, 'stroke-linecap': *}, label: {style: {'font-size': *, 'font-family', fill: (*|string)}, text: *}}}
 */
const getButtonOptions = (buttonSettings, chart, zoomEnabled) => {
    var btnBorderConf = buttonSettings.border,
        btnLabelConf = buttonSettings.label,
        showBtnBorder = btnBorderConf.show,
        btnRadius = showBtnBorder ? btnBorderConf.radius : 0,
        dashStyle = getDashArray(btnBorderConf.style, btnBorderConf.size),
        priorityArray = [btnLabelConf, chart.systemConf.chart, chart.systemConf.canvas, { fontColor: 'white', fontSize: 11 }, $ZCG],
        text = pick(btnLabelConf.text, zoomEnabled ? 'Reset Zoom' : 'Reset Brush');

    return {
        label: {
            text: text,
            style: {
                ...btnLabelConf,
                fill: btnLabelConf.fontColor || 'white',
                'font-size': btnLabelConf.fontSize,
                'font-family': getValByPriority(priorityArray, 'fontFamily'),
                'text-shadow': getValByPriority(priorityArray, 'textShadow')
            }
        },
        button: {
            radius: btnRadius,
            fill: buttonSettings.fillColor,
            'fill-opacity': buttonSettings.fillOpacity,
            stroke: btnBorderConf.color,
            'stroke-width': showBtnBorder ? btnBorderConf.size : 0,
            'stroke-dasharray': dashStyle,
            'stroke-linecap': getLineCap(btnBorderConf.lineCap, dashStyle),
            marginTop: btnLabelConf.marginTop,
            marginRight: btnLabelConf.marginRight,
            marginBottom: btnLabelConf.marginBottom,
            marginLeft: btnLabelConf.marginLeft,
            filter: svgBoxShadow(buttonSettings.shadow, chart)
        }
    };
};
