// @zohocharts/zc/src/behaviours/visual-effects/chart/controllers/sankey-zoom-scale.js
import { reset_c } from '../../../../events/visualEffects/controllers/reset_c';
import { defined } from '../../../../utils/zcutils';
import { d3_select } from '../../../../core/d3-constants';
import { getSvgTransform } from '../../../../utils/svg-transforms';

export function applyZoomToSankeyPlot(instanceMap, selectionManager, transform) {
    let x = transform.x.toFixed(6),
        y = transform.y.toFixed(6),
        k = transform.k.toFixed(6),
        chart = this.chart;

    chart.wrap.selectAll('g.sankey-chart-main').attr('transform', 'translate(' + x + ',' + y + ') scale(' + k + ' ' + k + ')');

    chart.dataObject.sankeyLayout.zoomTransform = {x, y, k};
    chart.sankeyZoomTransform = {x, y, k};
}

export function applyZoomToSankeyLabels(instanceMap, selectionManager, transform) {
    let chart = this.chart,
        x = transform.x.toFixed(6),
        y = transform.y.toFixed(6),
        k = transform.k.toFixed(6),
        labels = chart.staticPaneSVG
            .selectAll('g.labelClippathholder')
            .selectAll('g.subseries_0')
            .selectAll('.zc_label');
    labels
        .attr('transform', function(d) {
            let d3node = d3_select(this),
                _x,
                _y;
            if (defined(d3node.attr('translateX'))) {
                _x = parseFloat(d3node.attr('translateX'));
                _y = parseFloat(d3node.attr('translateY'));
            } else {
                let translate = getSvgTransform(this).translate;
                _x = parseFloat(translate[0]);
                _y = parseFloat(translate[1]);
                d3node.attrs({
                    translateX: _x,
                    translateY: _y
                });
            }
            _x = _x * k + parseFloat(x);
            _y = _y * k + parseFloat(y);
            return 'translate(' + _x + ',' + _y + ')';
        });
}

export function clearOldHighlightSankey() {
    let chart = this.chart,
        selectionManager = chart.eventHandler.eventManager.selectionManager;
    reset_c.call(this, null, selectionManager);
    chart.lastPoint = null;
    selectionManager.nearest = null;
}

export function clearCacheForEvents(){
    this.chart.cache.search = {};
    this.chart.cache.data = {};
}
