// @zohocharts/zc/src/label/overlapHandler/point-travel.js
import { TOP_ALIGN, RIGHT_ALIGN, BOTTOM_ALIGN, LEFT_ALIGN } from '../../core/align-constants';
import { STACKLABELS_LABELTYPE, DATALABELS_LABELTYPE } from '../labeltype-constants';
import { UNDEFINED } from '../../core/zc-constants';
import { d3_ascending } from '../../core/d3-constants';
import { pick } from '../../utils/zcutils';
import { doOverlap } from '../../utils/geometry-utils';
import { isLabelWithinBound } from '../datalabelsUtils';
import { getPlotBound } from '../../plot/plotSize';

export function pointTravel(chartObj, array) {
    let dataObj = chartObj.dataObject;

    var isContinousChart = dataObj.isContinousChart,
        length = array.length,
        translateOrders = {
            top: ['hidden', 'topPlus', 'right', 'bottom', 'left'],
            bottom: ['hidden', 'right', 'left']
        };

    var currentPlotBound = getPlotBound(chartObj);
    var plotbound = {
        x1: currentPlotBound.left,
        y1: currentPlotBound.top,
        x2: currentPlotBound.right,
        y2: currentPlotBound.bottom
    };

    array = isContinousChart ? array.sort((a, b) => d3_ascending(a.x1, b.x1)) : array;
    for (var i = 1; i < length; i++) {
        var bool = 0,
            overlap = false,
            testCase = 0,
            factor = pick(array[i].factor, 1),
            radius = (array[i].pointRadius || 0) * factor,
            x1 = array[i].x1,
            y1 = array[i].y1,
            x2 = array[i].x2,
            y2 = array[i].y2,
            w = array[i].x2 - array[i].x1,
            h = array[i].y2 - array[i].y1,
            orient = pick(array[i].orient, 'top'),
            translateOrder = translateOrders[orient],
            previousLabels = array.slice(0, i),
            currentLabel,
            isHidden,
            newHookPosition;
        previousLabels = isContinousChart ? previousLabels.reverse() : previousLabels;

        do {
            //testCase 0 : actual position, testCase 1 : top, testCase 2 : right, testCase 3 : bottom, testCase 4 : left
            var currentStep = translateOrder[testCase];

            switch (currentStep) {
                case 'hidden':
                    currentLabel = array[i];
                    isHidden = currentLabel.overlap === 'hidden' || dataObj.isBubbleEnabled;
                    break;
                case 'topPlus':
                    (currentLabel.y1 = y1 - h / 2), (currentLabel.y2 = y2 - h / 2);
                    newHookPosition = BOTTOM_ALIGN;
                    break;
                case 'right':
                    currentLabel.x1 = (x1 + x2) / 2 + radius;
                    currentLabel.x2 = (x1 + x2) / 2 + radius + w;
                    currentLabel.y1 = y2 + 1;
                    currentLabel.y2 = y2 + h + 1;
                    newHookPosition = LEFT_ALIGN;
                    break;
                case 'bottom':
                    currentLabel.x1 = x1;
                    currentLabel.x2 = x2;
                    currentLabel.y1 = y2 + 2 * (radius + 1);
                    currentLabel.y2 = y2 + 2 * (radius + 1) + h;
                    newHookPosition = TOP_ALIGN;
                    break;
                case 'left':
                    currentLabel.x1 = (x1 + x2) / 2 - radius - w;
                    currentLabel.x2 = (x1 + x2) / 2 - radius;
                    currentLabel.y1 = y2 + 1;
                    currentLabel.y2 = y2 + h + 1;
                    newHookPosition = RIGHT_ALIGN;
                    break;
            }
            for (var index = i - 1; index >= 0; index--) {
                overlap = doOverlap(previousLabels[index], currentLabel);
                if (overlap) {
                    bool++;
                    break;
                }
            }
        } while (++testCase < 5 && overlap);
        if (bool != 0) {
            var isWithInPlot = dataObj.isPolarAxisCategory ? true : isLabelWithinBound(currentLabel, plotbound), //For web chart, within plot boundary check is not necessary
                translateFunc = array[i].translateFunction,
                id = array[i].id,
                isStackedLabels = array[i].isStackedLabels;

            if (!isWithInPlot || bool == 5 || isHidden) {
                translateFunc(id, null, null, true, UNDEFINED, UNDEFINED, isStackedLabels ? STACKLABELS_LABELTYPE : DATALABELS_LABELTYPE);
                array[i].x1 = array[i].y1 = array[i].x2 = array[i].y2 = 0;
            } else {
                var translateX = currentLabel.x1 - x1,
                    translateY = currentLabel.y1 - y1;
                translateFunc(
                    id,
                    translateX,
                    translateY,
                    null,
                    undefined,
                    newHookPosition,
                    isStackedLabels ? STACKLABELS_LABELTYPE : DATALABELS_LABELTYPE
                );
            }
        }
    }
}
