// @zohocharts/zc/src/label/overlapHandler/slot-travel-helpers.js
//$Id$
//overlappingHandler.js
import { mathAbs, mathFloor, mathMax } from '../../core/zc-constants';
import { doOverlap } from '../../utils/geometry-utils';

export function arrange(slot, TOP, slotCount, slottedCount) {
    var leftUnslotted = [],
        unslotted = [],
        remaining = [],
        remainingCount = slotCount - slottedCount,
        sortByDist = function(a, b) {
            return a.dist - b.dist;
        },
        sortByY = function(a, b) {
            return a[0].y1 - b[0].y1;
        };
    var remainingCount_copy = remainingCount;

    //Collect free slot indices
    for (var i = 0; i < slotCount; i++) {
        if (!slot[i]) {
            leftUnslotted.push(i);
        }
    }

    slot.forEach(function(d) {
        d.sort(sortByDist);
    });

    // Add all unslotted labels
    slot.forEach(function(d) {
        d.forEach(function(dt, i) {
            if (i != 0) {
                remaining[i] = remaining[i] || [];
                remaining[i].push(dt);
            }
        });
    });

    remaining.forEach(function(d) {
        d.sort(sortByDist);
    });

    remaining.forEach(function(d) {
        d.forEach(function(dt, i) {
            unslotted.push([dt]);
        });
    });

    unslotted.forEach(function(d, i) {
        if (remainingCount > 0) {
            remainingCount--;
        } else {
            d[0].hide = true;
        }
    });

    unslotted = unslotted.slice(0, remainingCount_copy);
    unslotted.sort(sortByY);

    unslotted.forEach(function(d) {
        var us = d[0];
        var height = us.y2 - us.y1;
        var co = mathMax(mathFloor((us.y1 - TOP) / height), 0);
        if (slot[co]) {
            var dist,
                pos = leftUnslotted[0],
                id = 0;
            for (var o = 0; o < leftUnslotted.length; o++) {
                var dst = mathAbs(TOP + leftUnslotted[o] * height - us.y1);
                if (dist == null) {
                    dist = dst;
                }
                if (dst < dist) {
                    dist = dst;
                    pos = leftUnslotted[o];
                    id = o;
                }
            }
            if (pos < co) {
                leftUnslotted.splice(id, 1);

                var prev = pos;
                slot.forEach(function(d, i) {
                    var dt = d[0];
                    if (i > pos && dt.y1 < us.y1) {
                        slot[prev] = d;
                        prev = i;
                    }
                });

                slot[prev] = [us];
            } else {
                leftUnslotted.splice(id, 1);

                var prev = [us];
                for (var i = co; i <= pos; i++) {
                    if (!slot[i]) {
                        slot[i] = prev;
                    } else if (slot[i][0].y1 > us.y1) {
                        var temp = slot[i];
                        slot[i] = prev;
                        prev = temp;
                    }
                }
            }
        } else {
            slot[co] = d;
        }
    });

    slot.forEach(function(d, i) {
        var dt = d[0];
        var height = dt.y2 - dt.y1;
        var dy = TOP + i * height - dt.y1;
        dt.translate = mathAbs(dy) < height && i + 1 < slotCount && !slot[i + 1] && dy < 0 ? 0 : dy;
    });
    // Some other way
    slot.forEach(function(d, i) {
        var dt = d[0];
        var height = dt.y2 - dt.y1;
        var dy = TOP + i * height - dt.y1;
        dt.translate =
            mathAbs(dy) < height && i + 1 < slotCount && (!slot[i + 1] || !slot[i + 1][0].translate) && dy < 0 ? dt.translate : dy;
    });
    // Some other way
}

export function addToSlot(half, count, slot, TOP) {
    half.forEach(function(obj) {
        var height = obj.y2 - obj.y1;
        var co = mathMax(0, mathFloor((obj.y1 - TOP) / (obj.y2 - obj.y1)));
        obj.dist = obj.y1 < TOP ? TOP - obj.y1 : obj.y1 - (TOP + co * height);
        if (!slot[co]) {
            slot[co] = [];
            count++;
        }
        slot[co].push(obj);
    });
    return count;
}

export function addToHalf(obj, slotCount, TOP, BOTTOM, prev, half, count, totalSpaceNeeded, reverse) {
    var MARGIN = 5;
    if (!obj.hide) {
        if (slotCount == null) {
            slotCount = mathFloor((BOTTOM - TOP) / (obj.y2 - obj.y1));
        }
        var dist = reverse // If (dist < 0), it is overlapping
            ? prev == null
                ? BOTTOM - obj.y2
                : half[prev].y1 - obj.y2
            : prev == null
            ? obj.y1 - TOP
            : obj.y1 - half[prev].y2;
        totalSpaceNeeded += prev != null && doOverlap(obj, half[prev]) && dist < -MARGIN ? mathAbs(dist) : 0;

        prev = count;
        half.push(obj);
        count++;
    }
    return {
        prev: prev,
        count: count,
        totalslot: slotCount,
        need: totalSpaceNeeded
    };
}
