// @zohocharts/zc/src/behaviours/behavioural-events-mappings.js
import { ZCChartScroll } from './scroll/zc-chart-scroll';

export const CHART_BEHAVIOURAL_EVENT_MAPPINGS = {
    PLOT_ZOOM: {
        _property: 'chart.zoom',
        _feature_name: 'zoom',
        _handlers: chart =>
            chart.isGeoMap ? '_mapZoomHandlers' : chart.dataObject.isFlowChartCategory ? '_sankeyZoomHandlers' : '_chartZoomHandlers',
        _behavior: {
            _constructor: 'ZoomD3',
            _arguments: '_getPlotZoomArguments'
        }
    },
    PLOT_BRUSH: {
        _property: 'chart.brush',
        _feature_name: 'brush',
        _handlers: chart => (chart.isGeoMap ? '_mapBrushHandlers' : '_chartBrushHandlers'),
        _behavior: {
            _constructor: 'BrushD3',
            _arguments: '_getPlotBrushArguments'
        }
    },
    CHART_SCROLL: {
        _property: 'chart.scroll',
        _feature_name: 'scroll',
        _handlers: '_chartScrollHandlers',
        _behavior: {
            _constructor: 'ZCChartScroll',
            _arguments: 'ZCChartScrollArguments'
        }
    }
};

export const LEGEND_BEHAVIOURAL_EVENT_MAPPINGS = {
    LEGEND_SIZE_SLIDER_BRUSH: {
        _property: 'legend.size',
        _feature_name: 'brush',
        _handlers: '_legendSizeSliderBrushHandlers',
        _id: 'z-0',
        _behavior: {
            _constructor: 'BrushD3',
            _arguments: '_getLegendSizeSliderBrushArguments'
        }
    },
    LEGEND_COLOR_BAND_BRUSH: {
        _property: 'legend',
        _feature_name: 'brush',
        _handlers: '_legendColorBandBrushHandlers',
        _id: 'clr-0',
        _behavior: {
            _constructor: 'BrushD3',
            _arguments: '_getLegendColorBandBrushArguments'
        }
    }
};
