// @zohocharts/zc/src/behaviours/behavioural-event-manager.js
import { each, invokeFunction, isArray, splat } from '../utils/zcutils';
import { Registry } from '../core/zc-registry';

export class BehaviouralEventManager {
    constructor(chart) {
        this.chart = chart;
        this.behaviours = {};
    }

    /**
     *
     * @param name
     * @param behaviour
     * @return {*}
     */
    getBehaviouralEventInstance(name, behaviour) {
        let manager = this,
            instanceHolder;

        if (behaviour) {
            let behaviourClassConstructor = Registry.getComponent(behaviour._constructor),
                args = Registry.getComponent(behaviour._arguments);

            instanceHolder = manager.behaviours[name] || {};
            instanceHolder.instance = new behaviourClassConstructor(args(manager.chart));
            manager.behaviours[name] = instanceHolder;
            return instanceHolder;
        }
    }

    dataFinder(behaviourInstance, finderMap) {
        behaviourInstance.dataFinder = finderMap;
    }

    setListenerTypes(behaviourInstance, listenerTypes) {
        behaviourInstance.listenerTypes = listenerTypes;
    }

    getData(behaviourInstance, eventType, mousePos) {
        let manager = this,
            finder = behaviourInstance.dataFinder,
            selectionManager = { point: null };

        if (finder[eventType]) {
            selectionManager.point = finder[eventType].apply(manager, [behaviourInstance, eventType, mousePos]);
        }
        return selectionManager;
    }

    attachListeners(behaviourInstance, listeners) {
        let manager = this;
        let listenerTypes = behaviourInstance.listenerTypes,
            instance = behaviourInstance.instance;

        each(listeners, (listeners, type) => {
            if (listenerTypes.has(type)) {
                splat(listeners).forEach(listener => {
                    instance.on(type, function() {
                        listener.apply(manager, [behaviourInstance, ...arguments]);
                    });
                });
            }
        });
    }

    /**
     *
     * @param behaviourEvent
     * @param type
     * @return {[]}
     */
    getBehaviours(behaviourEvent, type) {
        let behaviours = [],
            defaultBehaviours = behaviourEvent.defaultBehaviour[type],
            customBehaviours = behaviourEvent.customBehaviour && behaviourEvent.customBehaviour[type];

        if (customBehaviours) {
            behaviours = [...splat(customBehaviours).map(b0 => ({ type: 'custom', behaviour: b0 }))];
        }

        if (defaultBehaviours) {
            behaviours = [...behaviours, ...splat(defaultBehaviours).map(b0 => ({ type: 'default', behaviour: b0 }))];
        }

        return behaviours;
    }

    /**
     *
     * @param behaviourEvent
     * @param listeners
     * @param type
     * @param event
     */
    setBehaviours(behaviourEvent, listeners, type = 'default', event) {
        behaviourEvent[`${type}Behaviour`] = listeners;
    }

    associateBehaviours(behaviourEvent, type, point, mouse, event) {
        let manager = this,
            behaviours = manager.getBehaviours(behaviourEvent, type),
            allowDefault = true;

        if (isArray(behaviours) && behaviours.length) {
            behaviours.forEach(b => {
                if (b.type === 'custom') {
                    invokeFunction(true, b.behaviour, manager, behaviourEvent, point, mouse, event);
                    allowDefault = event.allowDefault;
                } else if (allowDefault) {
                    b.behaviour.apply(manager, [behaviourEvent, point, mouse, event]);
                }
            });
        }
    }

    detachListeners() {}
}
