// @zohocharts/zc/src/behaviours/behavioural-events.js
import { BehaviouralEventManager } from './behavioural-event-manager';
import { defined, each, getPropVal, getVal } from '../utils/zcutils';
import { Registry } from '../core/zc-registry';
import { setDefaultOptions } from '../core/defaulttheme';
import { CHART_BEHAVIOURAL_EVENT_MAPPINGS, LEGEND_BEHAVIOURAL_EVENT_MAPPINGS } from './behavioural-events-mappings';
import { noOperation } from '../core/zc-constants';

export class BehaviouralEvents {
    constructor(chart) {
        this.chart = chart;
        this.eventManager = new BehaviouralEventManager(chart);
        this.init();
    }

    init() {
        let behaviourEvent = this,
            chart = behaviourEvent.chart,
            _initBehaviour = (behaviourMap, name, map) => {
                behaviourEvent.initiateBehaviour(name, map);
            };

        each(CHART_BEHAVIOURAL_EVENT_MAPPINGS, _initBehaviour);
        each(LEGEND_BEHAVIOURAL_EVENT_MAPPINGS, _initBehaviour);

        each(this.eventManager.behaviours, behaviour => {
            if (behaviour && behaviour._init) {
                behaviour._init.call(this, behaviour);
            }
        });
    }

    update() {
        each(this.eventManager.behaviours, behaviour => {
            //#ZC2320 $behaviour &&
            if (behaviour && behaviour._update) {
                behaviour._update.call(this, behaviour);
            }
        });
        this.init();
    }

    initiateBehaviour(name, map = CHART_BEHAVIOURAL_EVENT_MAPPINGS) {
        let manager = this.eventManager,
            chart = manager.chart,
            systemConfig = chart.systemConf,
            behaviourMap = map[name],
            rootObject = getPropVal(systemConfig, behaviourMap._property) || {},
            customBehaviours = rootObject.actions,
            handlers = Registry.getComponent(getVal(behaviourMap._handlers, [chart])),
            isSupported = defined(handlers) && handlers !== noOperation && handlers.SUPPORTED(chart, behaviourMap._id);

        if (isSupported && rootObject.enabled) {
            // TODO CLEANUP [renderVisiblePointsOnly]
            // if any behaviour is enabled, we render only the visible points
            // for default scroll we render all the points
            let isChartBehaviour = map === CHART_BEHAVIOURAL_EVENT_MAPPINGS;
            if (!defined(chart.systemConf.chart.width) && !chart.renderVisiblePointsOnly && isChartBehaviour) {
                chart.renderVisiblePointsOnly = rootObject.renderVisiblePointsOnly !== false;
            }

            // set the options
            setDefaultOptions(behaviourMap._feature_name, systemConfig.chart, chart.globalProp);

            let behaviouralInstance = manager.getBehaviouralEventInstance(name, getVal(behaviourMap._behavior, [chart]));

            // checking a key
            if (handlers.BEHAVIOUR) {
                behaviouralInstance._init = handlers.INIT;
                behaviouralInstance._update = handlers.UPDATE;
                // set listeners and behaviours
                // can be initialized directly
                manager.setBehaviours(behaviouralInstance, handlers.BEHAVIOUR);
                manager.dataFinder(behaviouralInstance, handlers.DATA_FINDER);
                manager.setListenerTypes(behaviouralInstance, handlers.LISTENER_TYPES);
                manager.attachListeners(behaviouralInstance, handlers.LISTENER);

                // set the custom behaviour if defined
                if ((customBehaviours = rootObject.actions)) {
                    manager.setBehaviours(behaviouralInstance, customBehaviours, 'custom');
                }
            }
        } else if (defined(manager.behaviours[name])) {
            manager.behaviours[name].destroy && manager.behaviours[name].destroy();
            manager.behaviours[name] = null;
            delete manager.behaviours[name];
        }
    }
}

Registry.setComponent('BehaviouralEvents', BehaviouralEvents);
