// @zohocharts/zc/src/events/bullet-events.js
//bullet-events.js

import { invokeFunction, defined, getPropVal, pick, isArray } from '../utils/zcutils';
import { getFormattedValue, generateDateVal } from '../formatter/formatter';
import { d3_map, d3_ascending, d3_descending } from '../core/d3-constants';
import { pointInRectangle } from './events-utils';
import { TIME_DATATYPE } from '../core/datatype-constants';
import { getColumnIdxBasedOnDataindex } from './events-utils';
import { mathMin, mathMax } from '../core/zc-constants';
import { events } from '../core/main-zc-init';
import { HIGHLIGHT_CONTEXT } from './visualEffects/highlight/highlight-context';

/**
 *
 * @param x
 * @param y
 * @param width
 * @param height
 * @param axesRotated
 * @returns {{x: Number, y: Number, width: Number, height: Number}}
 */

function getDimension(x, y, width, height, axesRotated) {
    return {
        _x: axesRotated ? y : x,
        _y: axesRotated ? x : y,
        _width: axesRotated ? height : width,
        _height: axesRotated ? width : height
    };
}

/**
 *
 * @param chart
 * @param barOptions
 * @param scaleX
 * @param scaleY
 * @constructor
 */
function BulletEvent(chart, barOptions, scaleX, scaleY) {
    this.chart = chart;
    this.barOptions = barOptions;
    this.scale = {
        x: scaleX,
        y: scaleY
    };
    this.barOptions = barOptions;
    this.axesRotated = chart.axes.rotated;

    // cache to store rectangles
    this.cache = d3_map();
}

/**
 *
 * @param coordinate
 * @param point
 * @returns {*}
 */

BulletEvent.prototype.getCurrentPoint = function(coordinate, point) {
    var bulletEvent = this,
        rectangles = bulletEvent.getCurrentRectangles(point),
        count = rectangles.length,
        currentRect;

    while (count--) {
        currentRect = rectangles[count];
        if (pointInRectangle(coordinate.x, coordinate.y, currentRect)) {
            return currentRect;
        }
    }
    return null;
};

/**
 *
 * @param point
 * @returns {Array}
 */
BulletEvent.prototype.getCurrentRectangles = function(point) {
    if (point.isDummy) {
        return [];
    }

    var bulletEvent = this,
        chart = bulletEvent.chart,
        xKey = chart.dataset.getX(point);

    // check cache and if already calculated return that.
    if (bulletEvent.cache.has(xKey)) {
        return bulletEvent.cache.get(xKey);
    }

    // else continue calculate rectangles
    var scaleY = bulletEvent.scale.y,
        scaleX = bulletEvent.scale.x,
        levelMarker = chart.systemConf.chart.plot.plotoptions.bullet.levelMarker,
        targetMarker = chart.systemConf.chart.plot.plotoptions.bullet.targetMarker,
        barOption = bulletEvent.barOptions,
        x = scaleX(xKey) + barOption.levelMarkerPadding,
        y,
        bandWidth = scaleX.bandwidth(),
        width = mathMin(levelMarker.maxBandWidth, bandWidth),
        height,
        type = 'y',
        datatype = chart.dataset.getDataType(type, 0),
        formatConf,
        rectangles = [],
        originalValues = point[levelMarker.dataindex] || [],
        col,
        axisrotated = bulletEvent.axesRotated,
        itemIndex = scaleX.domain().indexOf(xKey),
        tempRect = null;

    if (!isArray(originalValues)) {
        originalValues = [originalValues];
    }

    var steps = originalValues.slice(0);

    if (datatype === TIME_DATATYPE) {
        var column = chart.dataset.getColumn(type, 0);
        formatConf = defined(column.time) ? column.time : null;
    }

    steps.push(0); // add for get the base.
    steps = steps.map(function(point) {
        return datatype === TIME_DATATYPE ? generateDateVal(point, null, formatConf).getTime() : point;
    });

    steps = axisrotated ? steps.sort(d3_ascending) : steps.sort(d3_descending);

    var temp = steps.map(function(target) {
            return scaleY(target);
        }),
        len;

    steps.splice(steps.indexOf(0), 1); // remove
    var eventOptions_levelMarker = levelMarker.events || {};
    var levelMarkerEventsEnabled = defined(eventOptions_levelMarker.enabled) ? eventOptions_levelMarker.enabled : true;
    col = temp.length - 1;
    len = col - 1;
    if (levelMarkerEventsEnabled) {
        while (col--) {
            height = temp[col + 1] - temp[col];
            if (height > 0) {
                y = temp[col];
            }

            y = temp[col];

            if (height > 0) {
                tempRect = getDimension(x, y, width, height, axisrotated);
                tempRect.currentStage = steps[col];
                tempRect.type = 'levelMarker';
                tempRect.data = point;
                tempRect.rangeIndex = len - col;
                tempRect.itemIndex = itemIndex;
                tempRect.seriesIndex = point.seriesIndex;
                tempRect.events = levelMarker.events;
                tempRect.chartType = chart.chartTypes[0];
                tempRect._defaultH = defaultH;
                tempRect._visible = true;
                tempRect._key = `levelMarker_${point.seriesIndex}_${itemIndex}_${tempRect.rangeIndex}`;
                rectangles.push(tempRect);
            }
        }
    }

    // get the marker
    var eventOptions = targetMarker.events || {};
    var targetEventsEnabled = defined(eventOptions.enabled) ? eventOptions.enabled : true;
    if (targetMarker.enabled === true && targetEventsEnabled) {
        // get the dimension
        var strokeWidthFactor = 0; //mathMax(targetMarker.outerStrokeWidth, targetMarker.innerStrokeWidth) / 2;

        width = pick(targetMarker.size, barOption.levelMarkerWidth - 2 * pick(targetMarker.padding, barOption.padding / 2));
        if (defined(targetMarker.size)) {
            width = mathMin(width, bandWidth / 10) * 10;
        }

        if (targetMarker.symbol === 'line') {
            height = mathMax(pick(targetMarker.outerStrokeWidth, targetMarker.innerStrokeWidth, 0));
        } else {
            height = width;
        }

        // get the coordinates
        var currentStage = point[targetMarker.dataindex];

        x = scaleX(chart.dataset.getX(point)) - strokeWidthFactor;
        y = scaleY(currentStage) - strokeWidthFactor;

        x += (bandWidth - width) / 2;
        y -= height / 2;
        if (height < 5) {
            height = 5;
            y -= 2.5;
        }

        // The target marker bound might be very small. triggeringOffset extends its boundary for event

        var triggerOffset = eventOptions.triggerOffset || 0;
        x -= triggerOffset / 2;
        y -= triggerOffset / 2;
        width += triggerOffset + strokeWidthFactor * 2;
        height += triggerOffset + strokeWidthFactor * 2;

        tempRect = getDimension(x, y, width, height, axisrotated);
        tempRect.type = 'marker';
        tempRect.data = point;
        tempRect.currentStage = currentStage;
        tempRect.itemIndex = itemIndex;
        tempRect.seriesIndex = point.seriesIndex;
        tempRect.targetIndex = 0; // we support only one marker for one bullet.
        tempRect.events = targetMarker.events;
        tempRect.chartType = chart.chartTypes[0];
        tempRect._defaultH = defaultH;
        tempRect._key = `targetMarker_${point.seriesIndex}_${itemIndex}_${currentStage}`;
        rectangles.push(tempRect);
        tempRect._visible = true;
    }

    // add to cache
    bulletEvent.cache.set(xKey, rectangles);
    return rectangles;
};

/**
 *
 * @param currentPoint
 * @param seriesIndex
 * @returns {*}
 */
BulletEvent.prototype.getTooltipContent = function(currentPoint, seriesIndex) {
    var bulletEvent = this,
        chart = bulletEvent.chart,
        plotOptions = chart.systemConf.chart.plot.plotoptions.bullet,
        levelMarkerColumnInfo = getColumnIdxBasedOnDataindex(chart, plotOptions.levelMarker),
        targetMarkerColumnInfo = getColumnIdxBasedOnDataindex(chart, plotOptions.targetMarker),
        levelMarkerEvents = getPropVal(plotOptions, 'levelMarker.events') || {},
        targetMarkerEvents = getPropVal(plotOptions, 'targetMarker.events') || {},
        columnInfo = (this.type === 'marker' ? targetMarkerColumnInfo : levelMarkerColumnInfo) || chart.dataset.getColumn('y', 0),
        value = currentPoint.currentStage,
        dataType = columnInfo.datatype,
        metaInfo = chart.dataset.getMetaInfoObj(chart, currentPoint.data, seriesIndex),
        formattedValue = getFormattedValue(columnInfo[dataType], value, dataType, chart, 'tooltip'),
        text;

    var userCallback = {
        value: value,
        formattedValue: formattedValue,
        actualData: currentPoint.data,
        data: metaInfo
    };

    if (currentPoint.type === 'marker') {
        text = invokeFunction(targetMarkerEvents.tooltipContent, userCallback, chart, 'tooltip') || `Target: ${formattedValue}`;
    } else {
        text = invokeFunction(levelMarkerEvents.tooltipContent, userCallback, chart, 'tooltip') || `Range: ${formattedValue}`;
    }

    return text;
};

function defaultH() {
    return [];
}
events.BulletEvent = BulletEvent;

export var _bulletEvents = events.BulletEvent;
