// @zohocharts/zc/src/events/chart-events.js
import { EventManager } from './event-manager';
import { tooltip_helpers } from '../tooltip/tooltip-helpers';
import {
    getEventConf,
    getEventObject,
    getTriggeringOffset,
    isDependentPointHighlightType,
    isBarChartType,
    isWithinPlotarea
} from './events-utils';
import { charttype, events, scale, chartcategory } from '../core/main-zc-init';
import { ORDINAL_DATATYPE } from '../core/datatype-constants';
import { getNearestPoint } from './dataSelection/nearest-data';
import { XY } from './events-utils';
import { getPlotDefaultBehaviour } from './visualEffects/behaviour-mapping';
import { _defaultPlotHandlers, __defaultPlotHandlers } from './listeners/handler-mapping';
import { d3_event } from '../core/d3-constants';
import { bbox, pick, validateNullVal, arrayIncludes, isArray, isObjectPropDefined, mergeJSON } from '../utils/zcutils';
// import { ReverseLegendHighlight } from './reverse-legend-highlight';
import { UnSelectedHighlight } from './unselected-highlight';
import { getMousePosition } from './eventActions';
import { DataProcessor } from '../core/data-processor';
import { triggerUserCallback } from './customCallbackUtils';
import { isFacet } from '../axis/hierarchy-axis-utils';
import { Registry } from '../core/zc-registry';
import { mathMax } from '../core/zc-constants';
import { getChartTypeByIndex, isAreaChartType } from '../core/charttype-helpers';
import { getNearestPaneID } from '../axis/hierarchy-events-helpers';
import { MARKER_COMPONENT } from '../components/component-name-constants';
import { _barPointSelection } from '../bar/bar-selection';
import { customGestureList } from './custom-gesture';

export class ZC_ChartEvents {
    constructor(chart, bindTo) {
        this.chart = chart;
        this.tracker = bindTo;  //! memory leak
        this.rotated = chart.axes.rotated;
        this.isAxisCategory = chart.isAxisCategory;
        this.isGeoMap = chart.isGeoMap;
        this.isPolarAxisCategory = chart.dataObject.isPolarAxisCategory;
        this.isFlowChartCategory = chart.dataObject.isFlowChartCategory;
    }
    init() {
        this.eventManager = new EventManager(this.chart);
        this.reverseLegendHighlight = new (Registry.getComponent('legend.ReverseLegendHighlight'))(this.chart);
        this.unselectedHighlight = new UnSelectedHighlight(this.chart);
        this.eventManager.bindTo(this.tracker);
        this.eventManager.init();
        if (this.chart.isGeoMap) {
            return; //TODO Soon
        }
        this.addListeners(this.tracker);
        this.eventManager.dataFinder({
            mousemove: ZC_ChartEvents.findData,
            click: ZC_ChartEvents.findData,
            doubleclick: ZC_ChartEvents.findData
        });
        // this.eventManager.selectionManager = new SelectionManager(this.chart);
        this.eventManager.defaultBehaviour = getPlotDefaultBehaviour();
        this.eventManager.customBehaviour = ZC_ChartEvents.getCustomBehaviour;
        this.eventManager.getBehaviourName = ZC_ChartEvents.getBehaviourName;

        if (events.customPoints) {
            this.CustomPoints = new events.customPoints(this.chart);
            this.CustomPoints.addPoints(this.chart.cache.thresholdCustomPoints, true);
        }
    }
    update() {
        this.rotated = this.chart.axes.rotated;
        this.isAxisCategory = this.chart.isAxisCategory;
        this.isGeoMap = this.chart.isGeoMap;
        this.isPolarAxisCategory = this.chart.dataObject.isPolarAxisCategory;
        this.isFlowChartCategory = this.chart.dataObject.isFlowChartCategory;
        this.tooltipElement.initialize();
        this.bulletEvents = null; //#ZC1009

        if (this.CustomPoints) {
            this.CustomPoints.addPoints(this.chart.cache.thresholdCustomPoints, true);
        }
    }
    isSingleSource() {
        let event = d3_event(),
            chart = this.chart;

        if (event && !this.isGeoMap && chart.dataObject.isCombinationalChart && isFacet(chart)) {
            if (chart.divScroll) {
                let mouse = getMousePosition(chart.eventHandler.trackerElement.node(), event),
                    accessor = chart.eventHandler.dataAccesor({ x: mouse[0], y: mouse[1] }),
                    paneId = accessor.paneId(),
                    chartType = getChartTypeByIndex(chart, paneId, 0);

                return !arrayIncludes(chartcategory.nonAxisSingleSourceEvent, chartType);
            }

            let target = event.target,
                eventRectNode = chart.eventHandler.trackerElement.node(),
                seriesHolderNode = chart.svg.selectAll('g.seriesholder').node();

            if (target !== eventRectNode && !seriesHolderNode.contains(target)) {
                let nearest = chart.eventHandler.eventManager.selectionManager.nearest,
                    _isArray = nearest && isArray(nearest);
                if (nearest && ((_isArray && nearest.length) || !_isArray)) {
                    return _isArray;
                }
            } else {
                return target === eventRectNode;
            }
        }
        if(chart.viewFlow && chart.viewFlow.spec.renderOldConfig == false){
            return false;
        }

        return (
            this.isAxisCategory ||
            this.isPolarAxisCategory ||
            this.isFlowChartCategory ||
            this.isGeoMap ||
            chart.dataObject.singleSourceEvent
        );
    }
    clearOldActions() {
        var tooltip = this.chart.eventHandler.tooltipElement;
        if (tooltip != null && tooltip.timeout != null) {
            clearTimeout(tooltip.timeout); // To avoid flickering on  mouse move, clearing the timeout that makes the tooltip's display prop to none
        }
    }
    getSourceNode() {
        let node;
        if (this.isSingleSource()) {
            node = this.chart.eventHandler.trackerElement.node();
        } else {
            node = this.chart.svg.node();
        }
        return node;
    }
    getRelativeMousePosition(node, relativeTo) {
        var mouse = getMousePosition(node, d3_event());
        if (relativeTo) {
            var element = bbox(node);
            mouse[0] -= element.x;
            mouse[1] -= element.y;
            if (this.isPolarAxisCategory) {
                mouse[0] -= this.chart.plotarea_approx.x;
                mouse[1] -= this.chart.plotarea_approx.y;
            }
        }
        this.chart.eventHandler.mousePos = mouse;
        return {
            x: mouse[0],
            y: mouse[1]
        };
    }
    toNewObject(obj) {
        let o = { nearest: obj };
        this.eventManager.updateSelectionManager(o);
        return this.eventManager.selectionManager;
    }
    addListeners(tracker, type) {
        var _this = this;
        var handlers = (this.handlers = this.handlers || __defaultPlotHandlers);

        let clickHandlers = {
                click: function(event, args){
                    handlers.click.call(_this, event, args)
                },
                dblclick: function(event, args){
                    handlers.doubleclick.call(_this, event, args)
                },
                contextmenu: function(event, args){
                    handlers.contextmenu.call(_this, event, args)
                }
            },
            moveHandlers = { 
                mousemove: function(event, args){
                    handlers.mousemove.call(_this, event, args)
                },
                mouseout: function(event, args){
                    handlers.mouseout.call(_this, event, args)
                } 
            },
            touchHandlers = {
                tap: function(event, args){
                    handlers.click.call(_this, event, args)
                },
                dbltap: function(event, args){
                    handlers.doubleclick.call(_this, event, args)
                },
                singletouchmove: function(event, args){
                    handlers.mousemove.call(_this, event, args)
                },
                touchend: function(event, args){
                    handlers.mouseout.call(_this, event, args)
                }
            },
            options = {
                target: 'plot',
                chartObj: this.chart
            };

        if (type == 'tooltip') {
            this.eventManager.attachListeners(tracker, moveHandlers, null, options);
        } else {
            this.eventManager.attachListeners(tracker, moveHandlers, null, options, 'chart');
            this.eventManager.attachListeners(tracker, clickHandlers, null, options, 'chartmove');
            this.eventManager.attachListeners(tracker, {}, touchHandlers, options);
        }
        if (customGestureList) {
            let gestureKeys = Object.keys(customGestureList),
                CustomGesture = Registry.getComponent('CustomGesture');
            gestureKeys.map(d => {
                CustomGesture.call(this, d, tracker, 'plot');
            });
        }
    }
    removeTracker() {
        var i = this.trackerElement;
        i.attr('transform', null);
        i.attr('width', null);
        i.attr('height', null);
        i.attr('style', null).attr('cursor', null);
        i = null;
    }

    dataAccesor(mouse) {
        var manager = this,
            chart = this.chart,
            containsBarchart = manager.chart.dataObject.noofBarEnabledSeries > 0,
            isPolar = chart.dataObject.isPolarAxisCategory,
            tooltipShared = tooltip_helpers.hasSharedSeries(chart),
            sharedCriteria = (chart.cache.search.bisect = containsBarchart || tooltipShared);
        var accessor = {
            chart: chart,
            parseKey: function(key) {
                if (chart.cache.thresholdCustomPoints && key.indexOf('threshold') !== -1) {
                    return 'threshold';
                }
            },
            bestChartType: function(key) {
                var cTypeBest = key == 'threshold' || key == 'segment' ? 'scatter' : key;
                return cTypeBest;
            },
            radius: function(key) {
                if (arrayIncludes(chartcategory.nonAxisSingleSourceEvent, charttype[key])) {
                    return mathMax(chart.plotarea.width, chart.plotarea.height);
                }

                if (chart.cache.thresholdCustomPoints && key.indexOf('threshold') !== -1 && chart.cache.thresholdCustomPoints[key]) {
                    return chart.cache.thresholdCustomPoints[key].triggerOffset;
                }
                if (key == MARKER_COMPONENT) {
                    return chart.dataObject.flattenPoints.get('pane0_0').get(key).size; // TODO: check
                }
                key = key == 'threshold' || key == 'segment' ? 'scatter' : key;
                var mSizeCheckNeededType = key == 'line' || key == 'scatter',
                    isSameSize4Series = mSizeCheckNeededType ? chart.dataObject.markerSize[key].size() == 1 : null,
                    markerSize = isSameSize4Series ? parseInt(chart.dataObject.markerSize[key].values()[0]) || 0 : 0,
                    subtype;

                if (key == 'bar_segment') {
                    key = 'scatter';
                    subtype = 'segment';
                    markerSize = 0.5;
                }

                var triggeringOffset = pick(getTriggeringOffset(chart, charttype[key], null, subtype), 10);

                return triggeringOffset + markerSize / 2;
            },
            sizeOffset: function(key) {
                if (arrayIncludes(chartcategory.nonAxisSingleSourceEvent, charttype[key])) {
                    return 0;
                }

                key = key == 'threshold' ? 'scatter' : key;
                let commonPlotOptions = chart.systemConf.chart.plot.plotoptions[key],
                    customTriggerOffset = commonPlotOptions && commonPlotOptions.events && commonPlotOptions.events.triggerOffset;
                return customTriggerOffset || 10;
            },
            extent: function() {
                return [chart.plotarea.width, chart.plotarea.height];
            },
            size: function(key) {
                var mSizeCheckNeededType = key == 'line' || key == 'scatter',
                    isSizeDiffers4Series = mSizeCheckNeededType ? chart.dataObject.markerSize[key].size() > 1 : null;

                return (
                    chart.dataObject.isBubbleEnabled ||
                    isSizeDiffers4Series ||
                    key == 'segment' ||
                    key == MARKER_COMPONENT ||
                    arrayIncludes(chartcategory.nonAxisSingleSourceEvent, charttype[key])
                ); //NEED TO CHECK
            },
            toArray: function(obj) {
                var xy = XY(this.rotated);
                return [obj[xy[0]], obj[xy[1]]];
            },
            bisectMode: function() {
                let datatype = chart.dataset.getDataType('x');
                return chart.dataObject.noofBarEnabledSeries || datatype == ORDINAL_DATATYPE ? 'discrete' : 'continous';
            },
            median: function() {
                let hasMultiLeaves, column;
                if (chart.dataObject.pivot) {
                    var hierarchical = chart.dataObject.hierarchical,
                        columnList = hierarchical.pivotLayout.getRootNodeList('column'),
                        columnLeavesCount = columnList.getLeavesCount();
                    hasMultiLeaves = columnLeavesCount > 1;
                    column = chart.cache.search.lastPane.split('_')[1];
                }
                var xaxisReversed = chart.systemConf.chart.axes.xaxis.reversed,
                    onlyBarCombo_noAreaChart =
                        chart.dataObject.noofBarEnabledSeries && !chart.dataObject.noofAreaEnabledSeries && tooltipShared,
                    xscale = hasMultiLeaves ? chart.axes.x.leaves[column].scale : chart.axes.x.scale,
                    datatype = this.chart.dataset.getDataType('x');
                var median = scale.invert(datatype, mouse, xscale, chart.axes.rotated, null, xaxisReversed, onlyBarCombo_noAreaChart, null, chart);
                median = datatype == ORDINAL_DATATYPE ? xscale.domain()[median] : median;
                return median;
            },
            spliceIndices: function() {
                return [];
            },
            paneIndex: function(row, column) {
                if (chart.dataObject.pivot) {
                    var panes = chart.dataObject.hierarchical.pivotLayout.getRootNodeList('layer').panes;
                    return panes[`${row}_${column}`].index;
                }
                return 0;
            },
            paneId: function() {
                return getNearestPaneID(chart, mouse);
            },

            hasDependantPoints: function() {
                return isDependentPointHighlightType(chart);
            },

            pointLiesInsidePlot: function(point) {
                var seriesindex = point.seriesIndex,
                    cType_Obj = getChartTypeByIndex(chart, point.paneid, seriesindex),
                    type = cType_Obj.value;
                if (
                    !isBarChartType(type) &&
                    !isAreaChartType(type) &&
                    !chart.dataObject.isPolarAxisCategory &&
                    !chart.dataObject.isFlowChartCategory
                ) {
                    return isWithinPlotarea(point, chart);
                }
                return true;
            },
            userSelection: tooltipShared && chart.systemConf.tooltip.sharedSeriesInfo.selection,
            isBisectMode: (key, defaultBisect) => {
                if (defaultBisect) {
                    return key != MARKER_COMPONENT && !isPolar;
                }
                return false;
            },
            hasSelection: function(point, mode) {
                if (mode == 'BISECT_QUADTREE') {
                    return _barPointSelection(point, mouse);
                }
            }
        };

        return accessor;
    }
    static findData(mouse) {
        let manager = this,
            chart = manager.chart,
            containsBarchart = manager.chart.dataObject.noofBarEnabledSeries > 0,
            tooltipShared = tooltip_helpers.hasSharedSeries(chart),
            sharedCriteria = (chart.cache.search.bisect = containsBarchart || tooltipShared),
            accessor = chart.eventHandler.dataAccesor(mouse),
            paneid = `pane${accessor.paneId()}`,
            flattendata1 = chart.dataObject.flattenPoints.get(paneid);
        chart.cache.search.lastPane = paneid;

        let chartTypes =
            isFacet(chart) &&
            chart.dataObject.hierarchical &&
            chart.dataObject.hierarchical.gridOptions[paneid.replace('pane', '')].chartTypes;
        if (chartTypes && arrayIncludes(chartcategory.nonAxisSingleSourceEvent, chartTypes[0].value)) {
            sharedCriteria = chartTypes.some(_ => isBarChartType(_.value));
        }

        var point = getNearestPoint(flattendata1, mouse, sharedCriteria, accessor);
        
        if(point.nearest && point.nearest.isParent && chart.eventHandler.parentEventDisabled(point.nearest)){
            point = [];
            return {
                nearby: tooltipShared ? point.nearby : null,
                nearest: point.nearest,
                dependents: point.dependents,
                surround: point.nearby
            }
        }

        return {
            nearby: tooltipShared ? point.nearby : null,
            nearest: point.nearest,
            dependents: point.dependents,
            surround: point.nearby
        };
    }
    static getCustomBehaviour(selectionManager, eventType) {
        let manager = this,
            chart = manager.chart,
            event = manager.targetEvent,
            selectedPoint = selectionManager.nearest || selectionManager.selectionSet[0] || selectionManager.oldSelectionSet[0],
            behaviour = [],
            eventConfig,
            extraparams;
        if (!selectedPoint && eventType !== 'contextmenu' && eventType != 'mouseout') {
            return;
        }
        if (eventType == 'contextmenu' || (!selectedPoint && eventType == 'mouseout')) {
            eventConfig = chart.systemConf.chart.plot.events[eventType];
        } else {
            var plotOptions, commonPlotOptions;
            var seriesindex = selectedPoint.seriesIndex,
                cType_Obj = getChartTypeByIndex(chart, selectedPoint.paneid, seriesindex),
                type = selectedPoint.type || cType_Obj.name;

            if (type === 'zc-marker') {
                commonPlotOptions = {
                    events: isObjectPropDefined(chart.cache, `parser.${[selectedPoint.__key]}.parser.event`, true) || {}
                };
            } else if (type == 'threshold-marker' || type == 'levelMarker' || type == 'marker') {
                commonPlotOptions = selectedPoint;
            } else if (selectedPoint._type == 'segment') {
                //for whiskers
                commonPlotOptions = chart.systemConf.chart.plot.plotoptions[cType_Obj.name].whiskers;
            } else {
                var seriesdata = DataProcessor.helpers.getSeriesdata(chart, selectedPoint.paneid, selectedPoint.seriesIndex || 0),
                    seriesIndex = selectedPoint.seriesIndex,
                    chartName = cType_Obj.name,
                    plotOptions = seriesdata.plotoptions,
                    commonPlotOptions = chart.systemConf.chart.plot.plotoptions[chartName],
                    evt = chart.zcEventType || chart.eventHandler.targetEvent;
                    if(selectedPoint.isParent){
                        commonPlotOptions = mergeJSON(commonPlotOptions.parent, commonPlotOptions);
                    }
            }
            (eventConfig = getEventConf(plotOptions, commonPlotOptions, chart, eventType)),
                (extraparams = {
                    x: selectedPoint._x,
                    y: selectedPoint._y
                });
        }

        if (eventConfig) {
            if (eventType == 'cursor') {
                return eventConfig;
            }
            behaviour.push({
                action: triggerUserCallback(eventConfig, selectedPoint, getEventObject(event), extraparams, chart),
                type: 'custom'
            });
            return behaviour;
        }
        return null;
    }

    static getBehaviourName(point) {
        let manager = this,
            chart = manager.chart;

        return point && isObjectPropDefined(chart.cache, `parser.${[point.__key]}.parser.event.behaviour`, true);
    }

    getPointType(p) {
        var seriesindex = p.seriesIndex,
            // type = p.type || this.chart.chartTypes[seriesindex].name;
            type = p.type || 'series';
        return type;
    }
    eventDisabled(d) {
        /*
         * returns true if the series is filtered or event is disabled for that series
         */
        var plotOptions = d.plotoptions;
        var cName = DataProcessor.helpers.getChartType(d.type || this.chart.userdata.seriesdata.type, 'string');
        var commonPlotOptions = this.chart.systemConf.chart.plot.plotoptions[cName];
        return d.disabled || validateNullVal(getEventConf(plotOptions, commonPlotOptions, this.chart, 'enabled'), true) == false;
    }
    parentEventDisabled(d){
        var plotOptions = d.plotoptions,
            cName = DataProcessor.helpers.getChartType(d.type || this.chart.userdata.seriesdata.type, 'string'),
            commonPlotOptions = this.chart.systemConf.chart.plot.plotoptions[cName],
            commonPlotOptions = mergeJSON(commonPlotOptions.parent, commonPlotOptions);
        return validateNullVal(getEventConf(plotOptions, commonPlotOptions, this.chart, 'enabled'), true) == false;
    }
    destroy() {
        this.tracker = this.trackerElement = null;  // memory leak fix
    }
}

Registry.setComponent('ZC_ChartEvents', ZC_ChartEvents);
