// @zohocharts/zc/src/events/customCallbackUtils.js
import { defined, getPropVal, isArray, mergeJSON, pick } from '../utils/zcutils';
import { SANKEY_CHARTNAME } from '../core/chartname-constants';

export function customCallback(mouse, selectionManager) {}
export function triggerUserCallback(callback, point, event, extraParams, chart) {
    var userArguments;

    // get arguments for threshold marker
    if (!point) {
        return {
            callback: callback
        };
    } else if (point.type === 'threshold-marker') {
        var key = point.thresholdType === 'x' ? 'x' : `y.groups[${point.yColumnIndex}]`;
        userArguments = getPropVal(chart.axes, `${key}.threshold.line`).getMetaInfo(point);
    } else if (chart.chartTypes[0].name === SANKEY_CHARTNAME && point._type === 'node') {
        userArguments = {
            point,
            seriesIndex: point.seriesIndex,
            subSeriesIndex: point.subSeriesIndex,
            data: chart.dataset.getMetaInfoObj(chart, point, point.seriesIndex)
        };
    } else if(point.type === "legend"){
        userArguments = {data: point.data, index: point.i}
    }
    else {
        userArguments = getUserArguments(chart, point);
    }

    // set extra params
    if (defined(extraParams) && typeof extraParams === 'object') {
        userArguments = mergeJSON(userArguments, extraParams, true, true);
    }

    return {
        args: userArguments,
        callback: callback
    };
}

/**
 *
 * @param point
 * @return {*}
 */
function getUserArguments(chart, point) {
    var params,
        rangePoint = point.currentQualitativeRangePoint,
        dataPoint = pick(point.point, point.data) || {},
        levels = pick(point.level, dataPoint && dataPoint.level);

    // it is a hierarchy chart.
    // indexes need to be parsed
    if (defined(levels)) {
        params = parseHierarchyLevel(levels);
    } else {
        params = {
            seriesIndex: pick(point.seriesIndex, dataPoint.seriesIndex),
            subSeriesIndex: pick(point.subSeriesIndex, point.subSeriesIndex, dataPoint.subSeriesIndex),
            itemIndex: pick(point.itemIndex, dataPoint.itemIndex) //TODO: priority should be corrected
        };
    }

    if (isArray(point.subseriesindex)) {
        point.subseriesindex.some(function(value) {
            if (defined(value)) {
                return (params.subSeriesIndex = value);
            }
        });
    }

    // handle bullet range points
    if (defined(rangePoint)) {
        params.rangeIndex = rangePoint.rangeIndex;
        params.targetIndex = rangePoint.targetIndex;
    }

    if (defined(params.seriesIndex) && defined(params.itemIndex)) {
        if (defined(dataPoint) && isArray(dataPoint.data)) {
            dataPoint = dataPoint.data;
        }

        if (!isArray(dataPoint)) {
            params.subSeriesIndex = pick(params.subSeriesIndex, 0);
            dataPoint = point;
            // dataPoint = getCurrentDataPoint(params.seriesIndex, params.subSeriesIndex, params.itemIndex, point.itemPosShared);

            // since all indexes has been set, set the label index alone.
            if (defined(dataPoint.level)) {
                params.levelIndices = parseHierarchyLevel(dataPoint.level).levelIndices;
            }
        }

        // TODO has to be revamp. Since we are passing all required params, It can be access through API.
        params.data = chart.dataset.getMetaInfoObj(chart, dataPoint, params.seriesIndex);
    }
    // pass the actual point
    params.point = dataPoint;

    return params;
}

function parseHierarchyLevel(level) {
    level = level.split('_');
    return {
        levelIndices: level.map(Number),
        seriesIndex: +level[0],
        itemIndex: +level[level.length - 1]
    };
}
