// @zohocharts/zc/src/events/custom-points.js
//custom-points.js
import { pick, defined, getPropVal, invokeFunction, splat, encodeHTML } from '../utils/zcutils';
import { getDistanceFromPoint, pointInRectangle, getTooltipPosition, getEventObject } from './events-utils';
import { NONE } from '../core/zc-constants';
import { mathMax } from '../core/zc-constants';
import { addSVGmarker } from '../utils/svg-symbols';
import { events } from '../core/main-zc-init';
import { d3_event } from '../core/d3-constants';

/**
 *
 * @param {Chart} chartInstance
 * @param {Array} points array of points
 * @param {Boolean} toCircle if point to be treated as circle
 * @constructor
 */

function CustomPoints(chartInstance, points, toCircle) {
    if (!(this instanceof CustomPoints)) {
        return new CustomPoints(points);
    }

    points = pick(points, []);

    if (toCircle) {
        points = CustomPoints.toCircle(points);
    }

    this.chart = chartInstance;
    this.points = CustomPoints.addType(points);
}

/**
 *
 * @param points
 * @param toCircle
 */
CustomPoints.prototype.addPoints = function(points, toCircle) {
    if (!points) {
        return;
    }
    for (var key in points) {
        if (points.hasOwnProperty(key)) {
            if (toCircle) {
                points[key] = CustomPoints.toCircle(points[key]);
            }
            Array.prototype.push.apply(this.points, CustomPoints.addType(splat(points[key]))); // adds new point to the existing list
        }
    }

    return this;
};

/**
 *
 * @param fromIndex
 * @param itemCount
 * @return {[]}
 */
CustomPoints.prototype.removePoints = function(fromIndex, itemCount) {
    return this.points.splice(fromIndex, itemCount || 1);
};

CustomPoints.prototype.hoverPoint = function(point, event, mouse, showTooltip, showMarker, forceHover) {
    showTooltip = showTooltip !== false;
    showMarker = showMarker !== false;

    if (showTooltip) {
        var node = this.chart.eventHandler.getSourceNode(),
            mouse = this.chart.eventHandler.getRelativeMousePosition(node, true);
        showTooltipOnPoint(this.chart, event || d3_event(), mouse, point, showMarker, forceHover);
    }
    return;
};

/**
 *
 * @param points
 * @return {*}
 */
CustomPoints.toCircle = function(points) {
    points.forEach(function(point) {
        if (!point._converted && !defined(point.r)) {
            var _data = point._data();
            point.cx = _data._x + point.width / 2; // center x
            point.cy = _data._y + point.height / 2; // center y
            point.r = mathMax(point.width, point.height) / 2; // radius
            point._converted = true; // set converted
        }
    });
    return points;
};

/**
 *
 * @param points
 * @return {*}
 */
CustomPoints.addType = function(points) {
    points.forEach(function(point) {
        var _data = point._data();
        if (defined(point.r)) {
            point._type = 'circle';
        } else if (defined(_data._x) && defined(_data._y) && defined(point.width) && defined(point.height)) {
            point._type = 'rect';
        } else {
            // not supported
            point._type = 'poly';
        }
    });
    return points;
};

/**
 * custom point tooltip
 *
 * @param chart
 * @param event
 * @param mouse
 * @param point
 * @param highlight
 */
function showTooltipOnPoint(chart, event, mouse, point, highlight, forceHover) {
    var oldInfo = chart.eventHandler.oldInfo,
        tooltip = chart.systemConf.tooltip,
        tooltipShared = chart.systemConf.tooltip.sharedSeriesInfo,
        wrap = chart.staticPaneSVG.selectAll('g.eventgroup'),
        tooltipEnabled = tooltip.enabled === true || forceHover === true,
        tooltipEvent = chart.eventHandler.tooltipElement,
        context;

    if (oldInfo && oldInfo.isNoInfo === false) {
        chart.eventHandler.mouseOut();
    }

    // hide all the highlighters
    clearTimeout(tooltipEvent.timeout);
    wrap.selectAll('#zcscatterhighlight,#zcbarhighlight,#zcareahighlight').style('display', NONE);
    var _data = point._data();
    var x = _data._x,
        y = _data._y;

    // fix mouse
    if (!defined(mouse)) {
        // relative;
        mouse = { x: x, y: y };
    }

    // fix event
    if (defined(event)) {
        event = getEventObject(event);
    }

    if (!defined(event)) {
        // mimic clientX and clientY
        var chartSVGPosition = chart.svg.node().getBoundingClientRect();

        event = {
            clientX: x + chart.plotarea.left + chartSVGPosition.x,
            clientY: y + chart.plotarea.top + chartSVGPosition.y
        };
    }

    // highlight the point based on type
    if (highlight) {
        // highlight for points
        if (point._type === 'circle') {
            context = showMarkerOnPoint(point, x, y, wrap, chart.plotarea, chart);
        }
    }

    if (tooltipEnabled) {
        var tooltipPos = getTooltipPosition(event, mouse, x, y, chart, null, chart.axes.rotated, tooltipShared, null, false, false),
            tooltipContent = [encodeHTML(point.text), point.formattedValue].join(': '),
            marker = point.marker,
            pointEvents = pick(point.events, {}),
            tooltipFunction = pointEvents.tooltipContent;

        if (defined(tooltipFunction)) {
            var temp = invokeFunction(
                tooltipFunction,
                event,
                {
                    data: [point.columnInfo],
                    itemIndex: point.itemIndex,
                    point: point.point,
                    pointFormatted: point.formattedValue
                },
                chart
            );
            if (defined(temp)) {
                tooltipContent = temp;
            }
        }

        tooltipPos.helpers.backgroundColor = pick(tooltip.backgroundColor, marker.fillColor, point.color);
        tooltipPos.helpers.borderColor = pick(tooltip.borderColor, marker.strokeColor, point.color);

        tooltipEvent.show(chart, tooltipPos, tooltipContent, false, tooltipPos.helpers);
    }
    return context;
}

/**
 * custom point marker for point
 *
 * @param point
 * @param x
 * @param y
 * @param wrap
 * @param plotArea
 * @param chartObj Chart
 */
function showMarkerOnPoint(point, x, y, wrap, plotArea, chartObj) {
    var scatterWrap = wrap
        .selectAll('#zcscatterhighlight')
        .style('display', null)
        .attr('transform', `translate(${plotArea.left},${plotArea.top})`);

    var hoverOptions = getPropVal(point.marker, 'hoveroptions') || {};
    var symbolOptions = {
        color: point.color,
        symbolSize: {
            inner: point._mSize / 2,
            outer: point._mSize
        },
        transform: [x, y]
    };
    addSVGmarker(scatterWrap, null, hoverOptions, symbolOptions, chartObj);
    return scatterWrap;
}

events.customPoints = CustomPoints;

export var _nonDataPlotEvents = events.customPoints;
