// @zohocharts/zc/src/events/document-events.js
//$Id$
//document-events.js
import { $ZCG } from '../core/zcglobal';
import { $Browser } from '../core/browser-detect';
import { d3_event, d3_select } from '../core/d3-constants';
import { isElementVisible } from '../utils/zcutils';
import { attachevents } from './eventActions';

var docEle = d3_select(document.documentElement);

export var docTapFunction = function(chartObj, point) {
    //For pivot pie,funnel, pyramid, sunburst, packedbubble chart reset the old highlighlight on the last pane interacted
    if (chartObj && $Browser.isTouchDevice && chartObj.dataObject.pivot && chartObj.cache.search.lastPane) {
        var nonAxisSingleSeries = !(chartObj.dataObject.isAxisCategory || chartObj.dataObject.isNonAxisMultiSeriesCategory);
        if (nonAxisSingleSeries && (point && point.paneid && point.paneid !== chartObj.cache.search.lastPane)) {
            $ZCG.clickArguments = []; //TODO
            resetOldHighlight(chartObj);
        }
    }
    if (
        $Browser.isTouchDevice &&
        ((!chartObj && $ZCG.clickArguments && $ZCG.previousChartObj) || ($ZCG.previousChartObj && $ZCG.previousChartObj != chartObj)) &&
        // ZC2736 & ZC2728
        $ZCG.previousChartObj && !$ZCG.previousChartObj.destroying
    ) {
        resetOldHighlight($ZCG.previousChartObj);
    }
};

function resetOldHighlight(chartInstance) {
    var container = chartInstance.container.node();
    // if (document.body.contains(container) && (container.style.display != NONE)) {
    if (isElementVisible(container)) {
        $ZCG.clickArguments[4] = true;

        var tooltipConfig = chartInstance.systemConf.tooltip;
        var tooltipHideDelay = tooltipConfig.hideDelay;
        tooltipConfig.hideDelay = 0; //Mouse out should happen without delay
        var nonAxisSingleSeries = !(chartInstance.dataObject.isAxisCategory || chartInstance.dataObject.isNonAxisMultiSeriesCategory);
        if (chartInstance.dataObject.pivot && nonAxisSingleSeries) {
            chartInstance.eventHandler.eventManager.selectionManager.oldSelectionSet = [];
        }

        chartInstance.eventHandler.tapOldInfo = null;
        chartInstance.eventHandler.handlers.mouseout.call(chartInstance.eventHandler);

        tooltipConfig.hideDelay = tooltipHideDelay; //reset the delay
    }
}
export var documentTapEvent = docTapFunction;

if ($Browser.isTouchDevice) {
    var callPlotMouseOut = function() {
        var now = new Date().getTime();
        if (!$ZCG.clickedTime || now - $ZCG.clickedTime > 500) {
            docTapFunction(d3_event().target.__chart);
        }
    };

    var docTapHandlers = {
        tap: callPlotMouseOut,
        taphold: callPlotMouseOut,
        dbltap: callPlotMouseOut
    };
    attachevents(docEle, {}, docTapHandlers, { preventDefault: false });

    docEle.on('touchmove', function(event) {
        docTapFunction(d3_event(event).target.__chart);
    });
}
