// @zohocharts/zc/src/events/event-manager.js
import { attachevents } from './eventActions';
import { SelectionManager } from './dataSelection/selection-manager';
import { getValue, defined, invokeFunction, isObject, pick, isString } from '../utils/zcutils';
import { getEventConf } from './events-utils';
import { SELECTED_NEW } from './dataSelection/selection-types';
import { getPlotDefaultBehaviour } from './visualEffects/behaviour-mapping';
import { BehaviourMap } from './visualEffects/behaviour/behaviour-map';

export class EventManager {
    constructor(chart, data) {
        this.chart = chart;
        this.data = data;
        this.selectionManager;
        this.handlers = {};
        this.dataGenerators = {};
        this.defaultBehaviour = {};
        this.customBehaviour = {};
        this.targetEvent = null;
    }
    init() {
        this.selectionManager = new SelectionManager(this.chart, this);
    }
    bindTo(element) {
        this.tracker = element;
    }
    attachListeners(tracker, mouseHandlers, touchHandlers, options, namespace) {
        attachevents(tracker, mouseHandlers, touchHandlers, options, namespace);
    }
    dataFinder(actions) {
        if (actions) {
            this.dataGenerators = actions;
        }
        return this.dataGenerators;
    }
    //founddata should be like
    // {
    //     'nearby': point.nearby,
    //     'nearest': point.nearest,
    //     'type': 'series'
    // }
    getData(eventType, mouse, args = []) {
        let finder = this.dataFinder()[eventType],
            foundData = finder.call(this, mouse, eventType, ...args);

        this.updateSelectionManager(foundData);
        return this.selectionManager;
    }

    updateSelectionManager(selection) {
        if (selection.dependents) {
            this.selectionManager.add([selection.nearest]);
            this.selectionManager.selectionSet = selection.dependents;

            if (this.selectionManager.nearest !== selection.nearest) {
                this.selectionManager.updateID([selection.nearest], SELECTED_NEW);
            }
        } else if (selection.nearby) {
            this.selectionManager.add(selection.nearby);
        } else if (selection.nearest) {
            this.selectionManager.add([selection.nearest]);
        } else {
            this.selectionManager.remove();
        }
        this.selectionManager.nearest = selection.nearest;
        this.selectionManager.surround = selection.surround;
    }

    associateBehaviours(eventType, selectionManager, mouse) {
        var behaviour = this.getbehaviour(eventType, selectionManager);
        var iscustom = behaviour[0]?.type == 'custom';

        if (iscustom) {
            behaviour = this.getCustomBehaviour(behaviour.shift(), behaviour);
        }

        behaviour.forEach(d => {
            if (d) {
                let allowifcustom = getValue(d.allowifcustom, this);
                if (d.type == 'custom') {
                    //custom events
                    iscustom = true;
                    var action = d.action;
                    invokeFunction(action.callback, this.targetEvent, action.args, this.chart);
                } else if (this.targetEvent?.allowDefault || d.allowDefault) {
                    //add extra default behaviours
                    if (iscustom) {
                        if (!allowifcustom) {
                            invokeAction(d, this, mouse, selectionManager);
                        }
                    } else {
                        invokeAction(d, this, mouse, selectionManager);
                    }
                } else if (iscustom && allowifcustom) {
                    //extra behaviour added only for custom events
                    invokeAction(d, this, mouse, selectionManager);
                } else if (!iscustom && !allowifcustom) {
                    //default behaviours
                    invokeAction(d, this, mouse, selectionManager);
                }
            }
        });
    }
    getbehaviour(eventType, selectionManager) {
        var behaviour = [],
            point = selectionManager.nearest;

        var _b = this.getBehaviourName && this.getBehaviourName(point);
        var _defaultB = pick(getPlotDefaultBehaviour(_b), this.defaultBehaviour)[eventType];
        var _customB = this.customBehaviour(selectionManager, eventType);
        if (defined(_customB)) {
            behaviour = behaviour.concat(_customB); //{'callback': '<>', args:{}}
            behaviour = behaviour.concat(_defaultB);
        } else {
            behaviour = behaviour.concat(_defaultB);
        }

        return behaviour;
    }
    /**
     * 
     * @param {import('./visualEffects/behaviour/behaviour-map').behaviour} custom
     * @param {import('./visualEffects/behaviour/behaviour-map').behaviour[]} defaults 
     */
    getCustomBehaviour(custom, defaults) {
        let action = custom.action;
        let defs = {
            behaviour: defaults,
            value: defaults.map(d => d.type),
            map: BehaviourMap
        };
        let behaviour = invokeFunction(action.callback, this.targetEvent, action.args, this.chart, defs);
        if (behaviour) {
            behaviour = behaviour.map(b => isString(b) ? ({ allowifcustom: true, ...BehaviourMap.get(b) }) : b);
        } else {
            behaviour = defaults
        }
        return behaviour;
    }

    /**
     * Checks if a user has allowed the fault function
     *
     * @param userFunction
     * @param event
     * @return {*|boolean}
     */
    isDefaultAllowed(userFunction, event) {
        return !defined(userFunction) || (userFunction && event.allowDefault === true);
    }
}

function invokeAction(d, eventManager, mouse, selectionManager) {
    let action = isObject(d) ? d.action() : d;
    return action.call(eventManager, mouse, selectionManager);
}
