// @zohocharts/zc/src/events/eventActions.js
//$Id$
//eventActions.js
import { Registry } from '../core/zc-registry';
import { $Browser } from '../core/browser-detect';
import { d3_select, d3_event } from '../core/d3-constants';
import { getEventConf } from './events-utils';
import { charttype } from '../core/main-zc-init';
import { defined, invokeFunction, bbox, isEmptyObject } from '../utils/zcutils';
import { geometryUtils } from '../utils/geometry-utils';
import { NONE, NULL } from '../core/zc-constants';
import { d3_mouse } from '../core/d3/mouse';

export const TOUCH_EVENTS = ['tap', 'taphold', 'dbltap', 'tapholdend', 'touchmove', 'touchend', 'singletouchmove', 'doubletouchmove'];

/**
 *
 * @param node
 * @return {Array}
 */
export function getMousePosition(node, event) {
    // For touch devices d3.mouse(node) will return incorrect values, if the document is scrolled. So that pageXOffset and pageYOffset is added.
    var mouse = d3_mouse(node, event);
    if (event.touches != null && !$Browser.ANDROID && $Browser.uiwebview && !$Browser.WKWebView) {
        //d3.mouse(node) considers pageXOffset and pageYOffset in android devices
        mouse[0] += window.pageXOffset;
        mouse[1] += window.pageYOffset;
    }
    return mouse;
}
Registry.setComponent('events.getMousePosition', getMousePosition);

export function getRelativeMousePosition(chartObj) {
    //use in zcevents
    var node = chartObj.eventHandler.trackerElement.node(),
        mouse = (chartObj.eventHandler.mousePos = getMousePosition(node, d3_event())),
        relativeelement = bbox(node),
        relMouseX = mouse[0] - relativeelement.x,
        relMouseY = mouse[1] - relativeelement.y;
    return {
        x: relMouseX,
        y: relMouseY
    };
}

/*
 * If mouse is on a rotated element, this returns the mouse based on rotation
 */
export function getRotatedMouse(mouse, point) {
    let { x, y, rotation } = point;

    if (defined(rotation)) {
        return geometryUtils.rotatePoint(mouse, rotation, { x, y });
    }

    return mouse;
}
/*
 *Events related to statechange
 */

export function stateChangeListener(chartObj, eventdata, ele, type, defaultCallback) {
    //type: chartinit, chartload, chartreload, axisinit, axisload, plotinit, plotload

    var callback = (eventdata && eventdata[type]) || defaultCallback;
    if (defined(callback)) {
        invokeFunction(callback, d3_event(), chartObj);
    }
}

export function attachevents(ele, mouseHandlers, touchHandlers, options, namespace) {
    /*
     * This function attaches event listeners to the element. For touch devices, events will be mapped with touchstart,touchmove and touchend.
     * ele -> To which element the listeners will be attached
     * namespace -> To register multiple listeners for the same event type, the optional namespace is used. (e.g) : "zoom" or null. Refer github.com/mbostock/d3/wiki/Selections#on
     * mouseHandlers -> An object of event type and listener. (e.g) : { "click" : function(){}, "mousemove" : function(){} }
     * touchleave -> whether to handle touch leave or not
     */
    options = options || {};
    namespace = options.namespace || namespace;
    if ($Browser.isMouseDevice && !isEmptyObject(mouseHandlers)) {
        attachMouseEvents(ele, namespace, mouseHandlers);
    }
    if ($Browser.isTouchDevice && !isEmptyObject(touchHandlers)) {
        attachTouchEvents(ele, namespace, touchHandlers, options);
    }
}

export function detachevent(ele, namespace, handlers) {
    namespace = namespace != null && namespace != '' ? `.${namespace}` : '';
    if ($Browser.isMouseDevice) {
        if (namespace !== '') {
            ele.on(namespace, null);
        } else {
            for (var attr in handlers) {
                ele.on(attr, null);
            }
        }
        handlers = null;
    }
}

function attachMouseEvents(ele, namespace, handlers) {
    if (ele.node() == null) {
        return;
    }
    // if (namespace) {
    namespace = `.${namespace}`;
    var updatedHandlers = {};
    for (var attr in handlers) {
        updatedHandlers[attr + namespace] = handlers[attr];
        ele.on(attr + namespace, updatedHandlers[attr + namespace]);
    }
    handlers = updatedHandlers;
    // }

    ele.on(handlers);
}

function attachTouchEvents(ele, namespace, handlers, options) {
    if (ele.node() == null) {
        return;
    }
    var touchleave = options.touchleave,
        target = options.target,
        chartObj = options.chartObj,
        plotEventOptions = (chartObj && chartObj.systemConf.chart.plot.events) || {},
        doPreventDefault = options.preventDefault !== false;
    var isFocusedTouchmove = !(target === 'plot' && plotEventOptions.isFocusedTouchmove == false);
    namespace = namespace != null ? `.${namespace}` : '';

    var MINDISTANCE = 30,
        SINGLETAPDURATION = 150,
        DOUBLETAPDURATION = 500,
        FOCUSEDMOVEDURATION = 100,
        TOUCHMOVETIMER = 50,
        TAPHOLDDURATION = 500;
    var prevElement,
        touchstart = {},
        lastMove = 0,
        lastTap = 0,
        tapDifference = DOUBLETAPDURATION - SINGLETAPDURATION,
        firstMove = null,
        isDoubletouchStarted = false,
        tapholdDetected,
        tapHoldTimer;

    var mousemoveFn = handlers.touchmove,
        touchendFn = handlers.touchend,
        tapFn = handlers.tap,
        dbltapFn = handlers.dbltap,
        singletouchmoveFn = handlers.singletouchmove,
        doubletouchmoveFn = handlers.doubletouchmove,
        tapholdFn,
        tapholdendFn;

    if (
        target === 'plot' &&
        getEventConf(null, chartObj.systemConf.chart.plot.plotoptions[chartObj.chartTypes[0].name], chartObj, 'taphold')
    ) {
        tapholdFn = handlers.tap;
        tapholdendFn = touchendFn;
    } else {
        tapholdFn = handlers.taphold;
        tapholdendFn = handlers.tapholdend;
    }

    var touchstartProp = `touchstart${namespace}`,
        touchmoveProp = `touchmove${namespace}`,
        touchendProp = `touchend${namespace}`;

    var updatedHandlers = {};
    updatedHandlers[touchstartProp] = function(event) {
        if (plotEventOptions.enabled == false) {
            return;
        }

        var touchObject = d3_event(event).touches;

        if (touchObject.length > 1) {
            return;
        }

        touchstart.time = new Date();
        touchstart.location = getMousePosition(this, event);
        firstMove = null;
        isDoubletouchStarted = false;
        tapholdDetected = false;

        if (tapholdFn) {
            var e = event;
            var arg = arguments;
            var self = this;
            tapHoldTimer = setTimeout(function() {
                d3_event(e);
                if (chartObj) {
                    chartObj.zcEventType = 'taphold'; //change
                }

                tapholdFn.apply(self, arg);

                tapholdDetected = true;
            }, TAPHOLDDURATION);
        }
    };
    updatedHandlers[touchmoveProp] = function(event) {
        if (plotEventOptions.enabled == false) {
            return;
        }

        var touchObject = d3_event(event).touches;

        if (tapholdFn) {
            var currentLocation = getMousePosition(this, event);
            var distance =
                touchstart.location != null
                    ? getEuclideanDistance(touchstart.location[0], touchstart.location[1], currentLocation[0], currentLocation[1])
                    : null;

            if (distance > MINDISTANCE) {
                clearTimeout(tapHoldTimer);
            } else {
                if (tapholdDetected) {
                    preventDefault();
                    return false;
                }
            }
        }

        if (touchObject.length > 1) {
            if (touchObject.length == 2 && doubletouchmoveFn) {
                isDoubletouchStarted = true;
                doubletouchmoveFn.apply(this, arguments);
                preventDefault();
            }
            return;
        }

        var currentTime = new Date().getTime();
        if (!firstMove) {
            firstMove = currentTime;
        }

        // Touchleave event is not available. So finding the element to which the touchmove would be called is necessary. Call the mouseout of previous element and mousemove of the current element
        if (touchleave && document.elementFromPoint != null) {
            var timer = lastMove == 0 || currentTime - lastMove > TOUCHMOVETIMER;
            var curTouch = touchObject[0];
            var currentEle = document.elementFromPoint(curTouch.pageX - window.pageXOffset, curTouch.pageY - window.pageYOffset);
            if (!timer || currentEle?.localName === 'svg' || currentEle?.id.match(/tooltip/gi) != null) {
                preventDefault();
                return;
            }

            if (currentEle != this) {
                // Call mouse out of the previous element
                if (prevElement != currentEle) {
                    var touchEndElement = prevElement || this;
                    var element = d3_select(touchEndElement);
                    var endEvent = element.on(touchendProp),
                        d = element.datum(),
                        identity = element.attr('id') || element.attr('class');
                    var i = identity != null ? identity.match(/(\d)+$/gi) : null;
                    i = i != null && i[0] != null ? parseInt(i[0]) : null;

                    if (endEvent != null && i != null) {
                        endEvent.call(touchEndElement, event, d);
                    }
                }
                // Call mouse move of the current element
                var element = d3_select(currentEle),
                    className = element.attr('class');
                if (chartObj.chartTypes[0].value === charttype.dial && (className === 'inner' || className === 'outer')) {
                    element = d3_select(currentEle.parentNode);
                }
                var moveEvent = element.on(touchmoveProp),
                    d = element.datum(),
                    identity = element.attr('id') || element.attr('class');
                var i = identity != null ? identity.match(/(\d)+$/gi) : null;
                i = i != null && i[0] != null ? parseInt(i[0]) : null;
                if (moveEvent != null && i != null) {
                    moveEvent.call(currentEle, event, d);
                    prevElement = currentEle;
                    lastMove = currentTime;
                }
                return;
            }
        }

        if (!(isFocusedTouchmove && firstMove - touchstart.time < FOCUSEDMOVEDURATION)) {
            if (singletouchmoveFn) {
                if (chartObj) {
                    chartObj.zcEventType = 'touchmove'; //change
                }

                singletouchmoveFn.apply(this, arguments);
                preventDefault();
            }
        }
    };
    updatedHandlers[touchendProp] = function(event, d) {
        if (plotEventOptions.enabled == false) {
            return;
        }

        var touchObject = event.touches;

        if (tapholdFn) {
            clearTimeout(tapHoldTimer);
            if (tapholdDetected) {
                tapholdDetected = false;
                preventDefault();
                if (tapholdendFn) {
                    tapholdendFn.apply(this, arguments);
                }
                return false;
            }
        }

        if (touchObject.length > 1) {
            return;
        }

        if (isDoubletouchStarted && doubletouchmoveFn && touchendFn) {
            isDoubletouchStarted = false;
            touchendFn.apply(this, arguments);
            preventDefault();
            return;
        }

        var currentTime = new Date(),
            currentLocation = getMousePosition(this, d3_event());
        var duration = currentTime - touchstart.time;
        var distance =
            touchstart.location != null
                ? getEuclideanDistance(touchstart.location[0], touchstart.location[1], currentLocation[0], currentLocation[1])
                : null;
        var tapLength = currentTime - lastTap;

        if (tapDifference > 0 && tapLength <= tapDifference && tapLength > 0 && distance <= MINDISTANCE) {
            // Double tap event
            if (dbltapFn) {
                if (chartObj) {
                    chartObj.zcEventType = 'doubletap'; //change
                }

                dbltapFn.apply(this, arguments);
                preventDefault();
            }
            lastTap = 0;
        } else if (duration <= SINGLETAPDURATION && distance <= MINDISTANCE) {
            // Single tap event
            if (tapFn) {
                if (chartObj) {
                    chartObj.zcEventType = 'tap'; //change
                }

                tapFn.apply(this, arguments);
                preventDefault();
            }
        } else {
            if (touchendFn) {
                touchendFn.apply(this, arguments);
            }
        }
        lastTap = currentTime;
    };

    // use capture only for brush events
    let useBubble = chartObj && chartObj.isAxisCategory && chartObj.systemConf.chart.brush.enabled;
    for (var attr in updatedHandlers) {
        ele.on(attr, updatedHandlers[attr], useBubble);
    }

    function getEuclideanDistance(x1, y1, x2, y2) {
        return Math.sqrt(Math.pow(x2 - x1, 2) + Math.pow(y2 - y1, 2));
    }

    function preventDefault() {
        let event = d3_event();
        if (doPreventDefault && event.cancelable) {
            event.preventDefault();
        }
    }
}

export function pointerEventsNonePolyfill(ele) {
    var callback = function() {
            ele.style('display', NONE);

            var bottomElement = document.elementFromPoint(d3_event().clientX, d3_event().clientY),
                bottomElementD3 = d3_select(bottomElement),
                currentEvent =
                    bottomElement &&
                    bottomElement.__on &&
                    bottomElement.__on.filter(function(e) {
                        return e.type === d3_event().type;
                    })[0].value;

            ele.style('display', NULL);
            d3_select(this).style('cursor', currentEvent ? bottomElementD3.style('cursor') : NULL);

            if (currentEvent) {
                currentEvent.call(bottomElement, bottomElementD3.datum(), bottomElementD3.attr('index') || 0);
            }
            return false;
        },
        mouseHandlers = {
            mousemove: callback,
            mouseout: callback,
            click: callback,
            dblclick: callback
        },
        touchHandlers = {
            singletouchmove: callback,
            touchend: callback,
            tap: callback,
            dbltap: callback
        };
    attachevents(ele, mouseHandlers, touchHandlers);
}
