// @zohocharts/zc/src/events/event-surface.js
//event-surface.js

import { inherit, extend, defined, splat } from '../utils/zcutils';
import { d3_select, d3_event } from '../core/d3-constants';
import { pointInCircle, getEventObject } from './events-utils';
import { attachevents, getMousePosition } from './eventActions';

var event_surface_rect_counter = 0,
    event_surface_circle_counter = 0;

/**
 *
 * @param {Element} element
 * @param {String} types
 * @param {String} namespace
 * @param {Function} handler
 */

function addEvent(element, types, namespace, handler) {
    types.split(' ').forEach(function(type) {
        d3_select(element).on(`${type}.${namespace}`, null);
        d3_select(element).on(`${type}.${namespace}`, handler);
    });
}

/**
 *
 * @param {Element} element
 * @param {String} type
 * @param {String} namespace
 */
function removeEvent(element, type, namespace) {
    d3_select(element).on(`${type}.${namespace}`, null);
}

/**
 * Check if an element is an array, and if not, make it into an array.
 */

/**
 *
 * @param {Number} x
 * @param {Number} y
 * @param {Object} surface
 * @returns {boolean}
 */

function pointInRectangle(x, y, surface) {
    return x >= surface.x && x <= surface.x + surface.width && y >= surface.y && y <= surface.y + surface.height;
}

/**
 * @param {String} type
 * @param {Object} currentPosition
 * @param {Object} surface
 * @returns {boolean}
 */

function withinBound(type, currentPosition, surface) {
    var x = currentPosition[0],
        y = currentPosition[1];
    switch (type) {
        case 'rect': {
            return pointInRectangle(x, y, surface);
        }
        case 'circle': {
            return pointInCircle(x, y, surface);
        }
    }
}

// mouse enter function
function mouseEnter(self, currentSurface) {
    dispatchEvent(self, 'mouseenter', currentSurface);
    self._dispachedMouseEnter = true;
    self._dispachedMouseLeave = false;
}

function mouseLeave(self) {
    self._dispachedMouseEnter = false;
    self._dispachedMouseLeave = true;
    dispatchEvent(self, 'mouseleave');
}

function forceMouseLeave(self) {
    if (defined(self._lastSurface) && !self._dispachedMouseLeave) {
        self._finalMouseLeave = true;
        mouseLeave(self);
    }
}

/**
 *
 * @param {RectangleEventSurface|CircleEventSurface} self
 * @param {String} type
 * @param {Function|undefined} callback
 */

function defaultHandler(self, type, callback) {
    var currentMousePosition = getMousePosition(self.element, d3_event()),
        currentSurface,
        insideBound,
        inCount = 0,
        count = self.surface.length;

    if (self.element.ownerSVGElement) {
        var relativeMouse = self.element.getBBox();
        currentMousePosition[0] -= relativeMouse.x;
        currentMousePosition[1] -= relativeMouse.y;
    }

    while (count--) {
        currentSurface = self.surface[count];
        insideBound = withinBound(self.type, currentMousePosition, currentSurface);

        if (insideBound) {
            if (type === 'move') {
                // when moving through overlapping surface,
                // then call mouseout from lastSurface and mouseenter in currentSurface

                if (defined(self._lastSurface) && self._lastSurface !== currentSurface) {
                    if (!self._dispachedMouseLeave) {
                        self._finalMouseLeave = false;
                        mouseLeave(self);
                    }
                    mouseEnter(self, currentSurface);
                }

                // Calls mouse enter for the first time
                // In order to work mouseenter and mouseleave correctly on
                // overlapping surface, use stopPropagation on mousemove

                if (self.callbacks.hasOwnProperty('mouseenter') && !self._dispachedMouseEnter) {
                    mouseEnter(self, currentSurface);
                }

                // Call mouse move every time.

                if (self.callbacks.hasOwnProperty('mousemove')) {
                    d3_event().preventDefault();
                    dispatchEvent(self, 'mousemove', currentSurface, currentMousePosition);
                }

                inCount++;
                self._dispachedMouseLeave = false;
            } else if (defined(type)) {
                dispatchEvent(self, type, currentSurface, currentMousePosition);
            }

            if (defined(callback) && typeof callback === 'function') {
                callback.call(currentSurface || self._lastSurface, getEventObject(d3_event()), currentMousePosition);
            }

            self._lastSurface = currentSurface;

            if (self._stopPropagation === true) {
                break; // stop, if stopPropagation was called while runtime/
            }
        }
    }
    return inCount > 0;
}

/**
 *
 * @param {RectangleEventSurface|CircleEventSurface} self
 * @param {Object|null} currentSurface
 * @param {String} eventType
 * @param {Array} currentMousePosition is an array of current mouse position corresponding to the element.
 */

function dispatchEvent(self, eventType, currentSurface, currentMousePosition) {
    if (self.callbacks.hasOwnProperty(eventType)) {
        self.callbacks[eventType].call(currentSurface || self._lastSurface, getEventObject(d3_event()), currentMousePosition);
    }
}

function handleTouchEvents(self, element, options) {
    var touchHandlers = {
        tap: function() {
            self.trigger('click');
        },
        dbltap: function() {
            self.trigger('dblclick');
        },
        singletouchmove: function() {
            var insideBound = self.trigger('move');
            if (!insideBound && defined(self._lastSurface) && !self._dispachedMouseLeave) {
                self._finalMouseLeave = true;
                mouseLeave(self);
            }
            self._stopPropagation = false;
        },
        touchend: function() {
            forceMouseLeave(self);
        }
    };
    attachevents(d3_select(element), null, touchHandlers, options);
}

/**
 *
 * @param {Object} self is surface object
 * @param {Function|null} appHandler
 */

function handleMouseMoves(self, appHandler) {
    if (!self._addedMouseMove) {
        appHandler = defined(appHandler) ? appHandler : defaultHandler;
        addEvent(self.element, 'mousemove', self.namespace, function() {
            var insideBound = appHandler(self, 'move');
            if (!insideBound && defined(self._lastSurface) && !self._dispachedMouseLeave) {
                self._finalMouseLeave = true;
                mouseLeave(self);
            }
            self._stopPropagation = false;
        });

        self.callbacks.forceMouseLeave = forceMouseLeave;
        addEvent(self.element, 'mouseleave', self.namespace, function() {
            forceMouseLeave(self);
        });
    }
    self._addedMouseMove = true;
}

/**
 *
 * @param {Object} self is surface object
 * @param {Function|null} appHandler
 * @param {String} type mouse button events
 */

function handleClicks(self, appHandler, type) {
    appHandler = defined(appHandler) ? appHandler : defaultHandler;
    addEvent(self.element, type === 'click' ? 'click' : type, self.namespace, function() {
        appHandler(self, type);
        self._stopPropagation = false;
    });
}

// eventsListeners of EventSurface;

var eventsListeners = {
    /**
     *
     * @param {Function} callback
     * @param {Function|null} appHandler
     */

    click: function(callback, appHandler) {
        if (typeof callback === 'function') {
            this.callbacks.click = callback;
            this.callbacks.touchstart = callback;
            handleClicks(this, appHandler, 'click');
        }
    },

    /**
     *
     * @param {Function} callback
     * @param {Function|null} appHandler
     */

    dblclick: function(callback, appHandler) {
        if (typeof callback === 'function') {
            this.callbacks.dblclick = callback;
            handleClicks(this, appHandler, 'dblclick');
        }
    },

    /**
     *
     * @param {Function} callback
     * @param {Function|null} appHandler
     */

    mousedown: function(callback, appHandler) {
        if (typeof callback === 'function') {
            this.callbacks.mousedown = callback;
            handleClicks(this, appHandler, 'mousedown');
        }
    },

    /**
     *
     * @param {Function} callback
     * @param {Function|null} appHandler
     */

    mouseup: function(callback, appHandler) {
        if (typeof callback === 'function') {
            this.callbacks.mouseup = callback;
            handleClicks(this, appHandler, 'mouseup');
        }
    },

    /**
     *
     * @param {Function} callback
     * @param {Function|null} appHandler
     */

    contextmenu: function(callback, appHandler) {
        if (typeof callback === 'function') {
            this.callbacks.contextmenu = callback;
            handleClicks(this, appHandler, 'contextmenu');
        }
    },

    /**
     *
     * @param {Function} callback
     * @param {Function|null} appHandler is optional
     */

    mouseenter: function(callback, appHandler) {
        if (typeof callback === 'function') {
            this.callbacks.mouseenter = callback;
            handleMouseMoves(this, appHandler);
        }
    },

    /**
     *
     * @param {Function} callback
     * @param {Function|null} appHandler
     */

    mousemove: function(callback, appHandler) {
        if (typeof callback === 'function') {
            this.callbacks.mousemove = callback;
            this.callbacks.touchmove = callback;
            handleMouseMoves(this, appHandler);
        }
    },

    /**
     *
     * @param {Function} callback
     * @param {Function|null} appHandler is optional
     */

    mouseleave: function(callback, appHandler) {
        if (typeof callback === 'function') {
            this.callbacks.mouseleave = callback;
            handleMouseMoves(this, appHandler);
        }
    },

    /**
     *
     * @param {Object|Array} newSurface
     */

    update: function(newSurface) {
        this.surface = splat(newSurface);
    },

    stopPropagation: function() {
        this._stopPropagation = true;
    },

    remove: function() {
        for (var type in this.callbacks) {
            if (this.callbacks.hasOwnProperty(type)) {
                removeEvent(this.element, type, this.namespace);
            }
        }
    },

    /**
     *
     * @param {String} eventName event name
     * @param {Function|undefined} callback
     */

    trigger: function(eventName, callback) {
        return defaultHandler(this, eventName, callback);
    }
};

/**
 *
 * @param {Element} element
 * @param {Object|Array|null} surface
 * @constructor
 */

var EventSurface = function(element, surface) {
    this.element = element;
    this.surface = surface;
};

// add the methods to the prototype
extend(EventSurface.prototype, eventsListeners);

var init = function(element, surface, options) {
    this.events = [];
    this.callbacks = {};
    handleTouchEvents(this, element, options);
};

/**
 *
 * @param {Element} element
 * @param {Object} surface
 * @returns {RectangleEventSurface}
 * @constructor
 */

export var RectangleEventSurface = function(element, surface, options) {
    if (!defined(element) || element.nodeType !== 1 || !defined(surface)) {
        return;
    }

    if (!(this instanceof RectangleEventSurface)) {
        return new RectangleEventSurface(element, surface);
    }
    EventSurface.call(this, element, splat(surface));
    this.type = 'rect';
    this.namespace = `${this.type}Surface${event_surface_rect_counter++}`;
    init.call(this, element, surface, options);
};

// inherit EventSurface
inherit(RectangleEventSurface, EventSurface);

/**
 *
 * @param {Element} element
 * @param {Object} surface
 * @returns {CircleEventSurface}
 * @constructor
 */

export var CircleEventSurface = function(element, surface, options) {
    if (!defined(element) || element.nodeType !== 1 || !defined(surface)) {
        return;
    }

    if (!(this instanceof CircleEventSurface)) {
        return new CircleEventSurface(element, surface);
    }
    EventSurface.call(this, element, splat(surface));
    this.type = 'circle';
    this.namespace = `${this.type}Surface${event_surface_circle_counter++}`;
    init.call(this, element, surface, options);
};

// inherit EventSurface
inherit(CircleEventSurface, EventSurface);
