// @zohocharts/zc/src/events/events-utils.js
/**
 *
 * @param {Number} type chart type
 * @return {boolean}
 */
//events-utils.js
import { Registry } from '../core/zc-registry';
import { charttype, charttypenames, chartcategory } from '../core/main-zc-init';
import { SANKEY_CHARTNAME, WATERFALL_CHARTNAME } from '../core/chartname-constants';
import { geometryUtils } from '../utils/geometry-utils';
import { getSVGsymbol } from '../utils/svg-symbols';
import { toArray, isArray, defined, utils, isNumber, pick, arrayIncludes, isObjectPropDefined } from '../utils/zcutils';
import { NULL, NONE } from '../core/zc-constants';
import { mathSQRT, mathMin, mathMax } from '../core/zc-constants';
import { isDataActive } from '../core/data-processor/is-data-active';
import { colorscale_helpers } from '../axis/colorscale-helpers';
import { DataProcessor } from '../core/data-processor';
import { isFacet } from '../axis/hierarchy-axis-utils';
import { isSeriesFiltered, hasColorColumnInNonAxis } from '../utils/colorutils';
import { getChartTypeByIndex } from '../core/charttype-helpers';

export function getEventConf(plotOptions, commonPlotOptions, chartObj, eventType, bulletRect) {
    var events_spo = defined(plotOptions) && defined(plotOptions.events) && plotOptions.events, //spo - seriesplotoptions
        events_cpo = commonPlotOptions.events, //cpo - commonplotoptions
        events_plot = chartObj.systemConf.chart.plot.events;

    if (defined(bulletRect)) {
        events_cpo = commonPlotOptions[bulletRect.type] && commonPlotOptions[bulletRect.type].events;
    }

    var eventFn =
        defined(events_spo) && defined(events_spo[eventType])
            ? events_spo[eventType]
            : defined(events_cpo) && defined(events_cpo[eventType])
            ? events_cpo[eventType]
            : events_plot != null
            ? events_plot[eventType]
            : null;

    if (eventFn === undefined && events_plot) {
        eventFn = events_plot[eventType];
    }

    return eventFn;
}
Registry.setComponent('events.getConfig', getEventConf);

export function getEventObject(e) {
    return e.changedTouches ? e.changedTouches[0] : e;
}

/**
 *
 * @param {Number} type chart type
 * @return {boolean}
 */

export function isBarChartType(type) {
    return (
        type === charttype.bar ||
        type === charttype.stkdbar ||
        type === charttype.bullet ||
        type === charttype.gantt ||
        (isHeatMapChartType(type) && type !== charttype.geoheatmap) ||
        type === charttype.waterfall ||
        type === charttype.boxplot ||
        type === charttype.barrange
    );
}

/**
 * 
 * @param {number | string} type - chart number or chart name
 * @returns {boolean} defined in range chart or not
 */
export function isRangeChartType(type) {
    type = typeof type === 'string' ? charttype[type] : type;
    return chartcategory.rangePlot.has(type);
}

/**
 *
 * @param {Number} type chart type
 * @return {boolean}
 */

export function isScatterChartType(type) {
    return chartcategory.dotPlot.indexOf(type) != -1;
}

/**
 *
 * @param {Number} type chart type
 * @return {boolean}
 */

export function isBubbleChartType(type) {
    return chartcategory.bubble.has(type) || chartcategory.bubblePie.has(type);
}

/**
 *
 * @param {Number} type chart type
 * @return {boolean}
 */

export function isWaterfallChartType(type) {
    return seriesIs(type, WATERFALL_CHARTNAME);
}

/**
 *
 * @param {Number} type chart type
 * @return {boolean}
 */

export function isPointChartType(type) {
    return isScatterChartType(type) || type === charttype.bubblepie;
}

export function isSunburstChartType(type) {
    return type === charttype.sunburst;
}

export function isHeatMapChartType(type) {
    return type === charttype.heatmap || type === charttype.geoheatmap;
}

export function isPieChartType(type) {
    return chartcategory.piePlot.indexOf(type) != -1;
}

export function isMultiPieChartType(type) {
    return isPieChartType(type) && type !== charttype.pie;
}

/**
 *
 * @param type
 * @return {boolean}
 */
export function isGeoHeatMapChartType(type) {
    return type === charttype.geoheatmap;
}

export function seriesIs(type, condition) {
    var chartName = isNumber(type) ? charttypenames.get(type) : type;
    return chartName === condition;
}

/**
 *
 * @param {*} a
 * @param {*} b
 * @return {boolean}
 */

function equals(a, b) {
    return a === b;
}

/**
 *
 * @param {Object } chart current chart Object
 * @param {Array} point data point
 * @param seriesIndex series index of the corresponding data
 * @param commonOption commonOption for chart support
 * @return {{x, y, width, height}}
 */

export function getBarPosition(chart, point, seriesIndex, commonOption) {
    var getTransRectValue = function() {
        return chart.dataObject.getTransRectValue.apply(chart.dataObject, toArray(arguments));
    };
    return {
        x: getTransRectValue('x', point, seriesIndex, 0, commonOption),
        y: getTransRectValue('y', point, seriesIndex, 0, commonOption),
        width: getTransRectValue('width', point, seriesIndex, 0, commonOption),
        height: getTransRectValue('height', point, seriesIndex, 0, commonOption)
    };
}

/**
 *
 * @param {Object} chartObj
 * @param {Array} dataPoint
 * @param {Scale} xscale
 * @param {Scale} yscale
 * @param {Number} yaxiscolumnorder
 * @return {{x, y, width, height}}
 */

export function getHeatCellPosition(chartObj, dataPoint, xscale, yscale, yaxiscolumnorder) {
    var dataset = chartObj.dataset,
        commonPlotOptions = chartObj.systemConf.chart.plot.plotoptions.heatmap,
        symbolPath = commonPlotOptions.symbol && getSVGsymbol(commonPlotOptions.symbol),
        xValue = dataset.getX(dataPoint),
        yValue = dataset.getY(dataPoint, yaxiscolumnorder),
        zValue = chartObj.dataObject.hasZaxis ? chartObj.dataObject.getXYvalue(null, null, null, null, null, dataPoint, 'z') : 1,
        xBand = xscale.bandwidth(),
        yBand = yscale.bandwidth(),
        padding = commonPlotOptions.padding || 0,
        maxPaddingPercent = 0.4,
        xBandWithPadding = xBand - 2 * mathMin(padding, xBand * maxPaddingPercent),
        yBandWithPadding = yBand - 2 * mathMin(padding, yBand * maxPaddingPercent),
        x,
        y,
        w,
        h;

    if (chartObj.axes.rotated) {
        w = yBandWithPadding * zValue;
        h = xBandWithPadding * zValue;

        if (symbolPath) {
            w = h = mathMin(w, h);
        }

        y = xscale(xValue) + (xBand - h) / 2;
        x = yscale(yValue) + (yBand - w) / 2;
    } else {
        w = xBandWithPadding * zValue;
        h = yBandWithPadding * zValue;

        if (symbolPath) {
            w = h = mathMin(w, h);
        }

        x = xscale(xValue) + (xBand - w) / 2;
        y = yscale(yValue) + (yBand - h) / 2;
    }

    return {
        x: x,
        y: y,
        width: w,
        height: h
    };
}

/**
 *
 * @param {Array} dataPoint
 * @return {{x, y, width, height}}
 */

export function getPackedbubblePosition(dataPoint) {
    var w = 2 * dataPoint.r;
    var h = 2 * dataPoint.r;
    return {
        x: (dataPoint.x || 0) - w / 2,
        y: (dataPoint.y || 0) - h / 2,
        width: w,
        height: h
    };
}

export function crosshair_enabled(chartObj) {
    return chartObj.systemConf.tooltip.crosshair.type != NONE;
}

/**
 *
 * @param {Object} chartObj
 * @param {d3 selection} element
 * @param {Boolean} disabled
 */
export function setPointerEvents(chartObj, element, disabled, paneid) {
    if (!colorscale_helpers.isQuantileColorScale(chartObj) && !isFacet(chartObj)) {
        return;
    }

    var isPyramid = getChartTypeByIndex(chartObj, paneid, 0).value === charttype.pyramid,
        seriesdata_0 = hasColorColumnInNonAxis(chartObj)
            ? chartObj.dataObject.hierarchical.pivotLayout.getRootNodeList('layer').panes[paneid].data
            : isFacet(chartObj)
            ? DataProcessor.helpers.getSeriesdata(chartObj, paneid, 0).data[0]
            : isPyramid
            ? chartObj.seriesdata[0].data[0]
            : NULL,
        indexScale = colorscale_helpers.getIndexScale(chartObj);

    element.style('pointer-events', function(d, i) {
        var _ii = isPyramid ? seriesdata_0.length - 1 - i : pick(d && d.seriesIndex, i),
            dt = isPyramid && seriesdata_0 ? seriesdata_0[_ii] : d.data || d, //Pyramid hack
            isDisabled = utils.getVal(disabled, arguments),
            isFiltered;

        if (isFacet(chartObj)) {
            let _i = chartObj.dataObject.isHierarchy
                ? (d && d.seriesIndex) || (dt && dt.seriesIndex) || (dt && dt.seriesPos) || (dt && dt._seriesIndex) || 0 //TODO: check
                : hasColorColumnInNonAxis(chartObj)
                ? _ii
                : colorscale_helpers.getItemIndex(chartObj, indexScale, dt, _ii);

            isFiltered = isSeriesFiltered(chartObj, paneid, _i);
        }

        if (isDisabled || !isDataActive(isPyramid ? dt : d, chartObj) || isFiltered) {
            return NONE;
        }
    });
}

// get tooltip position for both ordinal and continuous scale
// TODO revamp continuous scale tooltip position.
export function getTooltipPosition(
    event,
    coordinates,
    x,
    y,
    chartObj,
    point,
    axisrotated,
    sharedTooltip,
    sharedTooltipInfo,
    isOrdinalChart,
    isPolar,
    fillColorFunc
) {
    var tooltipPos = null,
        plotarea = chartObj.plotarea,
        plotarea_approx = chartObj.plotarea_approx,
        // isSingleSourceEvent = chartObj.isAxisCategory || chartObj.dataObject.isPolarAxisCategory || chartObj.isGeoMap;
        isSingleSourceEvent = chartObj.eventHandler.isSingleSource(),
        yaxisorder = (point && point.yaxisOrder) || 0,
        axisreversed = isObjectPropDefined(chartObj.systemConf, `chart.axes.yaxis[${yaxisorder}].reversed`, true);

    point = point || {};

    if (sharedTooltip && crosshair_enabled(chartObj)) {
        if (axisrotated) {
            tooltipPos = {
                actLeft: x + plotarea.left,
                actTop: coordinates.y + plotarea.top,
                top: coordinates.y,
                left: isOrdinalChart ? plotarea.width : coordinates.x
            };
        } else {
            // common for both
            tooltipPos = {
                top: 0,
                actTop: plotarea.top + y
            };

            if (isOrdinalChart) {
                tooltipPos.left = coordinates.x;
                tooltipPos.actLeft = coordinates.x + plotarea.left;
            } else {
                tooltipPos.left = x;
                tooltipPos.actLeft = plotarea.left + x;
            }
        }
    } else {
        if (isOrdinalChart || !isSingleSourceEvent) {
            // common for both
            tooltipPos = {
                left: coordinates.x,
                top: coordinates.y
            };

            if (axisrotated) {
                var widthToAdd = axisreversed ? 0 : (point && point._width) || 0; //ZC2746 for bar chart with axisrotated x value with width should get add
                tooltipPos.actLeft = x + widthToAdd + plotarea.left;
                tooltipPos.actTop = coordinates.y + plotarea.top;
            } else {
                var heightToAdd = axisreversed ? (point && point._height) || 0 : 0; //ZC2746 for bar chart with axisrotated x value with width should get add
                tooltipPos.actLeft = coordinates.x + plotarea.left;
                tooltipPos.actTop = y + heightToAdd + plotarea.top;
            }
        } else {
            let chartType = getChartTypeByIndex(chartObj, point.paneid, point.seriesIndex || 0).value;

            if (chartObj.dataObject.isFlowChartCategory || arrayIncludes(chartcategory.nonAxisSingleSourceEvent, chartType)) {
                ({ x, y } = coordinates);
            }
            tooltipPos = {
                left: x,
                top: y,
                actLeft: plotarea.left + x,
                actTop: plotarea.top + y
            };
        }
    }
    var dLeft = coordinates.x - event.clientX,
        dTop = coordinates.y - event.clientY;

    tooltipPos.dLeft = dLeft + plotarea.left;
    tooltipPos.dTop = dTop + plotarea.top;
    tooltipPos.left -= dLeft;
    tooltipPos.top -= dTop;
    tooltipPos.distance = chartObj.dataObject.isBubbleEnabled && !isOrdinalChart && point._size ? point._size / 2 : 0; //DOUBT
    // tooltipPos.distance = chartObj.dataObject.getXYvalue(axisrotated, stacked(type), xscale, ysc, yscindex, point, 'z', chartObj.zscale) / 2;
    if (!isSingleSourceEvent) {
        tooltipPos.dLeft = dLeft;
        tooltipPos.dTop = dTop;
    }
    // tooltipPos.dLeft = dLeft + plotarea.left;
    // tooltipPos.dTop = dTop + plotarea.top;

    // add dependencies in helper variable

    tooltipPos.helpers = {
        seriesindex: point.seriesIndex,
        subseriespos: isArray(point.subSeriesIndex) ? point.subSeriesIndex[point.seriesIndex] : point.subSeriesIndex,
        subseriesLength: isArray(point.subseriesLength) ? point.subseriesLength[point.seriesIndex] : point.subseriesLength,
        itempos: pick(point.categoryIndex, point.sortedItemIndex, point.itemIndex),
        paneid: point.paneid,
        itemposShared: point.itemPosShared,
        color: chartObj.seriesColor,
        tooltipSharedInfo: sharedTooltipInfo,
        tooltipShared: sharedTooltip,
        svgFillFunc: fillColorFunc || (chartObj.renderer[point.seriesIndex] && chartObj.renderer[point.seriesIndex].fillColorFunc),
        inheritColors: chartObj.dataObject.isWaterfallEnabled,
        d: chartObj.dataObject.isHierarchy || !isSingleSourceEvent ? (point.data && point.data.data) || point.data || point : point,
        renderTo: 'plot',
        isWithinPlot:
            isOrdinalChart || !isSingleSourceEvent || chartObj.dataObject.isPolarAxisCategory
                ? true
                : pointInRectangle(chartObj.plotarea.left + x, chartObj.plotarea.top + y, {
                      x: chartObj.plotarea.left + (chartObj.scrolledX || 0),
                      y: chartObj.plotarea.top + (chartObj.scrolledY || 0),
                      width: chartObj.plotarea.width,
                      height: chartObj.plotarea.height
                  })
    };

    if (isPolar) {
        tooltipPos.dLeft += plotarea_approx.x;
        tooltipPos.dTop += plotarea_approx.y;
    }

    return tooltipPos;
}

export function getColumnIdxBasedOnDataindex(chartObj, options) {
    //check for multi yaxis
    var columns = chartObj.userdata.metadata.columns,
        columnidx = columns.filter(function(d) {
            return d.dataindex === options.dataindex;
        });
    return columnidx[0];
}

/**
 *
 * @param {Number} x
 * @param {Number} y
 * @param {Object} rectangle
 * @returns {boolean}
 */

export function pointInRectangle(x, y, rectangle) {
    return geometryUtils.pointInRectangle({ x: x, y: y }, rectangle);
}

/**
 *
 * @param {Number} x
 * @param {Number} y
 * @param {Object} surface
 * @returns {boolean}
 */

export function pointInCircle(x, y, surface) {
    return mathSQRT((x - surface.cx) * (x - surface.cx) + (y - surface.cy) * (y - surface.cy)) <= surface.r;
}

/**
 * returns distance from a point
 *
 * @param {Number} x mouse x
 * @param {Number} y mouse y
 * @param {{cx:Number, cy:Number,r: Number}} point
 */
export function getDistanceFromPoint(x, y, point) {
    return mathSQRT((x - point.cx) * (x - point.cx) + (y - point.cy) * (y - point.cy)) - (point.r || 0);
}

/**
 * returns distance between point and rectangle
 *
 * @param {{x:Number, y: Number, width: Number, height:Number}} rect
 * @param {{x:Number, y:Number}} point
 */
export function distanceBetweenRectFromPoint(rect, point) {
    var dx = mathMax(rect.x - point.x, 0, point.x - (rect.x + rect.width));
    var dy = mathMax(rect.y - point.y, 0, point.y - (rect.y + rect.height));
    return mathSQRT(dx * dx + dy * dy);
}

/**
 *
 * @param {Chart} chartInstance
 * @param {Number} chartType
 * @param {Number} bubbleTriggerOffset
 * @return {Number|null}
 */
export function getTriggeringOffset(chartInstance, cType, bubbleTriggerOffset, subtype) {
    var triggerOffset = null,
        commonPlotOptions = chartInstance.systemConf.chart.plot.plotoptions[charttypenames.get(cType)],
        isContinuousPointType = isPointChartType(cType) || cType === charttype.line, //without area,as area should have no mindist
        filledChartType = !isContinuousPointType,
        customTriggerOffset = commonPlotOptions && commonPlotOptions.events && commonPlotOptions.events.triggerOffset;

    if (defined(customTriggerOffset)) {
        triggerOffset = customTriggerOffset;
    } else if (isBubbleChartType(cType) || cType == 'bubble') {
        if (arguments.length === 3) {
            triggerOffset = arguments[2]; // assign as it is
        } else {
            var r = chartInstance.zscale ? chartInstance.zscale.range()[1] / 2 : 10,
                offset = chartInstance.dataObject.isCombinationalChart ? 10 : 50;
            triggerOffset = r + offset; // for bubble charts
        }
    } else if (filledChartType || chartInstance.systemConf.tooltip.sharedSeriesInfo) {
        triggerOffset = null; //for area charts and tooltip shared
    } else if ((chartInstance.dataObject.isCombinationalChart && isContinuousPointType) || subtype == 'segment') {
        //for whsiker segment
        triggerOffset = 10; // for combo chart with line
    } else {
        triggerOffset = 50; //for line charts
    }
    return triggerOffset;
}

export function isPointVisible(chart, point) {
    var isWithinClrRange = isDataActive(point, chart),
        isValidData = !isNullData(chart, point),
        isWithinCategories = !point._filter;
    return isWithinClrRange && isValidData && isWithinCategories;
}

export function isWithinPlotarea(point, chartObj) {
    let x1 = chartObj.plotarea.left + (chartObj.scrolledX || 0),
        y1 = chartObj.plotarea.top + (chartObj.scrolledY || 0),
        x2 = x1 + chartObj.plotarea.width,
        y2 = y1 + chartObj.plotarea.height,
        pointX = point._x + chartObj.plotarea.left,
        pointY = point._y + chartObj.plotarea.top;

    var condition = x1 <= pointX && pointX <= x2 && y1 <= pointY && pointY <= y2;
    if (condition) {
        return 1;
    } else {
        return null;
    }
}

function isNullData(chart, point) {
    let dataset = chart.dataset,
        dataProcessorHelpers = DataProcessor.helpers,
        isCatNullX = dataProcessorHelpers.isCatNull(dataset, 'x'),
        isCatNullY = dataProcessorHelpers.isCatNull(dataset, 'y', point.yaxisOrder);
    if (chart.seriesdata[0].type == charttype.stkdarea) {
        return false;
    }
    if (point) {
        // FEATURES_PIVOT_2 treemap null data hover
        let chartType = getChartTypeByIndex(chart, point.paneid, point.seriesindex).value;
        if (arrayIncludes(chartcategory.nonAxisSingleSourceEvent, chartType)) {
            return false;
        }
    }

    return point == null || isCatNullX(dataset.getX(point)) || isCatNullY(dataset.getY(point, point.yaxisOrder));
}

export function XY(axisrotated) {
    var X = 'x',
        Y = 'y';
    if (axisrotated) {
        X = 'y';
        Y = 'x';
    }
    return [X, Y];
}

/**
 *
 * @param {Chart|GeoMap} chart
 * @return {boolean}
 */
export function isDependentPointHighlightType(chart) {
    return chart.chartTypes[0].name === SANKEY_CHARTNAME;
}
