// @zohocharts/zc/src/events/reverse-legend-highlight.js
//$Id$
//reverse-legend-highlight.js
// import { legend_helpers } from '../legend/legend-helper';
import { legend_helpers } from '../core/main-zc-init';
import { NULL, NONE } from '../core/zc-constants';
import { d3_event } from '../core/d3-constants';
import { defined, jsonStringify, each, isEmptyObject, arrayIncludes } from '../utils/zcutils';
import { colorscale_helpers } from '../axis/colorscale-helpers';
import { getLegendView } from '../legend/events/legend-event-helpers';
import { getHoverOptions } from '../legend/events/legend-event-helpers';
import { SANKEY_CHARTNAME } from '../core/chartname-constants';
import { isBubbleChartType } from '../core/charttype-helpers';
import { Registry } from '../core/zc-registry';

export var ReverseLegendHighlight = function(chartObj, options) {
    var element = chartObj.legendarea.element,
        legendSeries = element && element.selectAll('.legseries, .zc-legend-legseries').node();

    this.chartObj = chartObj;
    this.legendSeries = legendSeries;
    this.legendElements = legendSeries ? legendSeries.children : [];
    this.itemHighlighted = false;
    this.hoveredSeriesIndex = NULL;
    this.hoveredlegendOrder = NULL;
    this.previousHoveredSeriesIndex = NULL;
    this.legend_series_height = NULL;
    this.legendScrollable = NULL;
    this.legendComponents = initComponents(chartObj.legends);

    if (legendSeries) {
        this.legend_series_height = legendSeries.clientHeight;
        this.legendScrollable = legendSeries.scrollHeight > this.legend_series_height;
    }
};
Registry.setComponent('legend.ReverseLegendHighlight', ReverseLegendHighlight);

ReverseLegendHighlight.prototype.highlight = function(options) {
    var chartObj = this.chartObj;

    // Size legend highlight ---
    each(this.legendComponents, (comp, id) => {
        let reverseEventManager = chartObj.legendEvents.reverseEventManager,
            reverseManager = reverseEventManager[getManagerID(chartObj, id)],
            legend = chartObj.legends.get(id);

        if (!reverseManager || isEmptyObject(reverseManager)) {
            // TODO: remove reverseManager != null
            return;
        }

        let { encode } = legend.meta,
            hoverOptions = getHoverOptions(reverseManager.options.config, encode, false);
        if (!hoverOptions.showOnDataHover || !options) {
            // options = null for sharedSeriesInfo
            return;
        }

        let dt =
                chartObj.isGeoMap && options.type
                    ? options.point
                    : encode === 'clr'
                    ? (options.data && options.data.data) || options.data || options // TODO: check
                    : options,
            previousHighlightedData = comp.data,
            highlightedData = reverseManager.getData('mousemove', dt);
        reverseManager.targetEvent = d3_event();

        if (previousHighlightedData && jsonStringify(highlightedData.point) !== jsonStringify(previousHighlightedData.point)) {
            reverseManager.associateBehaviours('mouseout', previousHighlightedData);
        }

        if (!hasEncode(encode, chartObj, options)) {
            return;
        }

        reverseManager.associateBehaviours('mousemove', highlightedData);

        comp.data = { ...highlightedData };
        comp.highlighted = true;
    });

    // --- Size legend highlight

    if (legend_helpers.isContinousLegend(chartObj)) {
        return;
    }

    let previousHoveredSeriesIndex = this.previousHoveredSeriesIndex,
        legendOptions = chartObj.systemConf.legend,
        hoverOptions = getHoverOptions(legendOptions, 'clr', false),
        legendComponent = chartObj.legend.component,
        commonRendererProp = options ? chartObj.renderer[options.rendererIndex || 0].commonRendererProp() : NULL;

    // get proper legend component based on legend order
    if (chartObj.dataObject.pivot && options) {
        var legendOrder = commonRendererProp.legendOrder();
        this.hoveredlegendOrder = legendOrder;
        legendComponent = chartObj.legends[legendOrder].component.discrete_encode;
    }

    if (!legendComponent) {
        return;
    }

    if (
        hoverOptions &&
        hoverOptions.showOnDataHover &&
        !(legend_helpers.isContinousLegend(chartObj) && (!hoverOptions.type || hoverOptions.type === NONE))
    ) {
        var oldInfo = chartObj.eventHandler.eventManager.selectionManager,
            highlightItem = legendComponent.highlightItem;
        if (
            !(
                chartObj.isAxisCategory ||
                chartObj.dataObject.isPolarAxisCategory ||
                chartObj.isGeoMap ||
                chartObj.dataObject.isFlowChartCategory
            ) ||
            (oldInfo && oldInfo.nearest)
        ) {
            var hoveredData, seriesIndex, itemIndex;
            if (
                chartObj.isAxisCategory ||
                chartObj.dataObject.isPolarAxisCategory ||
                chartObj.isGeoMap ||
                chartObj.dataObject.isFlowChartCategory
            ) {
                hoveredData = chartObj.isGeoMap && options.type ? options.point : options;
                seriesIndex = options.seriesIndex;
                itemIndex = options.itemIndex;
                if (chartObj.dataObject.pivot && options && hoveredData.type != 'threshold-marker') {
                    seriesIndex = commonRendererProp.legendIndex(); //proper series index for pivot chart with multi cluster rows
                }
            } else {
                hoveredData = options.data ? options.data.data || options.data : options;
                seriesIndex = options.seriesIndex;
                itemIndex = options.itemIndex;
                if (chartObj.dataObject.pivot && options) {
                    if (chartObj.dataObject.isHierarchy) {
                        var gridOptions = chartObj.dataObject.hierarchical.gridOptions;
                        seriesIndex = gridOptions[options.paneid || '0_0'].legendIndices[options.seriesIndex || 0];
                    } else {
                        seriesIndex = commonRendererProp.legendIndex();
                    }
                }
            }
            //#ZC1565
            var hoveredSeriesIndex = (this.hoveredSeriesIndex = this.getLegendIndex(
                hoveredData,
                chartObj.isAxisCategory ||
                    chartObj.dataObject.isPolarAxisCategory ||
                    chartObj.isNonAxisMultiSeriesCategory ||
                    chartObj.isGeoMap
                    ? seriesIndex
                    : itemIndex
            ));

            if (legend_helpers.isContinousLegend(chartObj)) {
                var clrValue = chartObj.dataset.getClr(hoveredData),
                    userRanges = legendOptions.colorBand.ranges,
                    brushScale = chartObj.axes.clr.brushScale,
                    pos;

                if (userRanges && userRanges.length) {
                    pos = brushScale(userRanges[hoveredSeriesIndex]) + brushScale.bandwidth() / 2;
                } else {
                    pos = brushScale(clrValue);
                }

                highlightItem(pos, true, NULL, userRanges && userRanges.length ? hoveredSeriesIndex : [clrValue]);
                this.itemHighlighted = true;
            } else {
                var legendElements = this.legendElements;
                if (legendElements[hoveredSeriesIndex]) {
                    var legendData = legend_helpers.getLegendData(chartObj);

                    if (hoveredSeriesIndex !== previousHoveredSeriesIndex && previousHoveredSeriesIndex != null) {
                        highlightItem(
                            legendElements[previousHoveredSeriesIndex],
                            legendData[previousHoveredSeriesIndex],
                            previousHoveredSeriesIndex,
                            legendOptions
                        );
                    }

                    highlightItem(
                        legendElements[hoveredSeriesIndex],
                        legendData[hoveredSeriesIndex],
                        hoveredSeriesIndex,
                        legendOptions,
                        true
                    );
                    this.itemHighlighted = true;

                    // scroll into view;
                    if (this.legendScrollable && !(chartObj.systemConf.legend.expandable && chartObj.systemConf.legend.expandable.show)) {
                        this.legendSeries.scrollTop = legendElements[hoveredSeriesIndex].offsetTop - this.legend_series_height / 2;
                    }
                }
            }
            this.previousHoveredSeriesIndex = hoveredSeriesIndex;
        }
    }
};

ReverseLegendHighlight.prototype.reset = function() {
    let chartObj = this.chartObj;

    // Size legend reset ---
    each(this.legendComponents, (comp, id) => {
        if (comp.highlighted) {
            let reverseEventManager = chartObj.legendEvents.reverseEventManager,
                previousHighlightedData = comp.data,
                reverseManager = reverseEventManager[getManagerID(chartObj, id)];

            if (!reverseManager || isEmptyObject(reverseManager)) {
                return;
            }

            reverseManager.targetEvent = d3_event();
            reverseManager.associateBehaviours('mouseout', previousHighlightedData);
            comp.data = null;
            comp.highlighted = false;
        }
    });

    // --- Size legend reset

    if (legend_helpers.isContinousLegend(chartObj)) {
        return;
    }

    if (this.itemHighlighted) {
        var legendElements = this.legendElements,
            hoveredSeriesIndex = this.hoveredSeriesIndex,
            legendComponent = chartObj.legend.component;

        if (chartObj.dataObject.pivot) {
            legendComponent = chartObj.legends[this.hoveredlegendOrder || 0].component.discrete_encode;
        }

        if (!legendComponent) {
            return;
        }

        let highlightItem = legendComponent.highlightItem;

        if (legend_helpers.isContinousLegend(chartObj)) {
            highlightItem(NULL, false);
        } else if (defined(legendElements[hoveredSeriesIndex])) {
            var legendData = legend_helpers.getLegendData(chartObj),
                legendOptions = chartObj.systemConf.legend;
            highlightItem(legendElements[hoveredSeriesIndex], legendData[hoveredSeriesIndex], hoveredSeriesIndex, legendOptions);
        }
        this.hoveredSeriesIndex = null;
    }
    this.itemHighlighted = false;
};

ReverseLegendHighlight.prototype.getLegendIndex = function(d, index) {
    var chartObj = this.chartObj,
        clrOptions = chartObj.axes.clr,
        isSankey = chartObj.chartTypes[0].name === SANKEY_CHARTNAME;

    if (clrOptions.scaleType) {
        return colorscale_helpers.getItemIndex(chartObj, clrOptions.indexScale, d);
    } else if (isSankey) {
        let renderer = chartObj.renderer[0],
            node = d._type === 'link' ? renderer.getLinkByPoint(d).source : renderer.getSankeyNodeByPoint(d);
        return node.point.itemIndex;
    } else if (index != null) {
        return index;
    }
};

function getManagerID(chart, id) {
    let legend = chart.legends.get(id);

    if (!legend) {
        null;
    }

    let { encode } = legend.meta,
        view = getLegendView(chart.legends, id);

    return `${fullEncode(encode).toUpperCase()}_${view.toUpperCase()}(${id})`;
}

function initComponents(legends = []) {
    let components = {},
        _init = legend => (components[legend.id] = { data: null, highlighted: false });

    legends.forEach(_init);
    return components;
}

function fullEncode(encode) {
    let list = ['shape'];
    return arrayIncludes(list, encode) ? encode : encode === 'z' ? 'size' : encode === 'clr' ? 'color' : null;
}

function hasEncode(encode, chartObj, options) {
    let { seriesIndex } = options;

    switch (encode) {
        case 'z':
            return defined(seriesIndex) && isBubbleChartType(chartObj.chartTypes[seriesIndex].value);
        case 'clr':
            return legend_helpers.isContinousLegend(chartObj);
    }
}

export { getManagerID };
