// @zohocharts/zc/src/events/unselected-highlight.js
//$Id$
import { jsonStringify } from '../utils/zcutils';
import { NULL, UNDEFINED } from '../core/zc-constants';
import { d3_select } from '../core/d3-constants';

export var UnSelectedHighlight = function(chartObj) {
    this.chartObj = chartObj;
    this.elements =
        chartObj.plot.renderer.mode === 'svg' ? chartObj.container.selectAll('g.seriesholder, g.qualitativeGroupHolder') : d3_select(chartObj.context.canvas);
    this.backupStyles = NULL;
    this.previousPoint = NULL;
};

UnSelectedHighlight.prototype.highlight = function(point) {
    var chartObj = this.chartObj,
        replacerFunction = chartObj.dataObject.isHierarchy ? replacer : NULL,
        applyUnselectedSeriesToPlot = chartObj.systemConf.legend.highlightEffect.applyUnselectedSeriesToPlot,
        previousPoint = this.previousPoint,
        condition =
            applyUnselectedSeriesToPlot &&
            point &&
            (!previousPoint ||
                jsonStringify(previousPoint.data, replacerFunction) != jsonStringify(point.data, replacerFunction) ||
                previousPoint.seriesPos != point.seriesPos ||
                previousPoint.itemPos != point.itemPos ||
                jsonStringify(previousPoint.subseriesindex) != jsonStringify(point.subseriesindex));

    if (condition) {
        var elements = this.elements,
            highlightEffect = chartObj.legendHighlight.getLegendHighlightEffect(0, 'unselectedSeries', 'plot', point.paneid),
            styles = chartObj.legendHighlight.getEffect(highlightEffect, NULL, NULL, NULL, 0, 'plot');

        this.getBackup(styles);
        elements.styles(styles);
        this.previousPoint = point;
    }
};

UnSelectedHighlight.prototype.reset = function() {
    var backupStyles = this.backupStyles;
    if (backupStyles) {
        this.elements.styles(backupStyles);
        this.backupStyles = NULL;
        this.previousPoint = NULL;
    }
};

UnSelectedHighlight.prototype.getBackup = function(styles) {
    var elements = this.elements,
        backupStyles = this.backupStyles;

    if (!backupStyles) {
        backupStyles = {};
        for (var styleName in styles) {
            if (styles.hasOwnProperty(styleName)) {
                backupStyles[styleName] = elements.style(styleName);
            }
        }
        this.backupStyles = backupStyles;
    }
};

function replacer(key, value) {
    // Filtering out properties
    var omitKeysList = ['parent', 'children'];
    if (omitKeysList.indexOf(key) > -1) {
        return UNDEFINED;
    }
    return value;
}
