// @zohocharts/zc/src/events/zcevents.js
//$Id$
//zcevents.js

import { events, charttypenames, charttype, HIGHLIGHT_EFFECT_REGEX, UNHIGHLIGHT_EFFECT_REGEX, zmaps, legend_helpers as L_H } from '../core/main-zc-init';
import { $ZCG } from '../core/zcglobal';
import {
    utils,
    isArray,
    pick,
    defined,
    mergeJSON,
    bbox,
    invokeFunction,
    validateNullVal,
    getPropVal,
    perToPx
} from '../utils/zcutils';
import { DataProcessor } from '../core/data-processor';
import { appendEle, createElement, roundedRect } from '../utils/zcsvgutils';
import {
    getEventConf,
    getEventObject,
    isBarChartType,
    isBubbleChartType,
    crosshair_enabled,
    getBarPosition,
    isHeatMapChartType,
    isMultiPieChartType,
    isPointChartType,
    getTriggeringOffset,
    getHeatCellPosition,
    getTooltipPosition
} from './events-utils';
import { isAreaChartType } from '../core/charttype-helpers';
import { documentTapEvent } from './document-events';
import { ORDINAL_DATATYPE, NUMERIC_DATATYPE } from '../core/datatype-constants';
import {
    d3_select,
    d3_min,
    d3_max,
    d3_curves,
    d3_bisector,
    d3_merge,
    d3_map,
    d3_lineRadial,
    d3_areaRadial,
    d3_area,
    d3_event
} from '../core/d3-constants';
import { scale } from '../core/main-zc-init';
import { CIRCLE_ELEMENT } from '../core/element-constants';
import { mathMin, mathMax, mathAbs, mathSQRT, mathAtan2, UNDEFINED } from '../core/zc-constants';
import { isAngleInside } from '../polaraxis/polarutils';
import { POINTER, NONE, NULL, TRANSPARENT } from '../core/zc-constants';
import { getSVGsymbol } from '../utils/svg-symbols';
import { isWithinColorRange, getRawColor, isMultiColoring } from '../utils/colorutils';
import { effects } from '../effects/zceffects';
// import { ReverseLegendHighlight } from './reverse-legend-highlight';
import { attachevents, getMousePosition, pointerEventsNonePolyfill } from './eventActions';
import { getBarCommonOption } from '../bar/barRendererConf';
import { tooltip_helpers } from '../tooltip/tooltip-helpers';
import { tooltip_content } from '../tooltip/tooltip-content-template';
import { annotation } from '../annotation/annotation';
import { addSVGmarker } from '../utils/svg-symbols';
import { bubblepieUtils } from '../bubblepie/bubblepie-utils';
// import { legend_helpers as L_H } from '../legend/legend-helper';
import { UnSelectedHighlight } from './unselected-highlight';
import { LOCATION_MARKERS } from '../geo-core/geo-marker-constant';
import { getBubblePieAngle } from '../bubblepie/bubblepiechart';
import { geometryUtils } from '../utils/geometry-utils';
import { Registry } from '../core/zc-registry';

events.mouseevents = function(
    chartObj,
    tooltipele,
    systemConf,
    plotarea,
    dataset,
    xscale,
    barscale1,
    barscale2,
    ysca,
    wrap,
    color,
    labelhighter
) {
    var element,
        gvpFn = utils.getValueOfPlotOptions;
    var datatype = dataset.getDataType('x');
    var yscale = ysca,
        color = color;
    var axisrotated = chartObj.axes.rotated,
        isPolarAxisCategory = chartObj.dataObject.isPolarAxisCategory,
        tooltipEnabled = systemConf.tooltip.enabled === true,
        defaultScrollElement = chartObj.container.selectAll('.zohocharts_scroll'),
        levelMarkerPoint,
        eventAttacherOptions = {
            target: 'plot',
            chartObj: chartObj
        },
        tapOldInfo,
        dataProcessorHelpers = DataProcessor.helpers,
        getParsedYAxisOrder = dataProcessorHelpers.getParsedYAxisOrder,
        getChartType = dataProcessorHelpers.getChartType,
        isCatNull = dataProcessorHelpers.isCatNull,
        tooltipShared = tooltip_helpers.hasSharedSeries(chartObj),
        bubblePieKeys = getAvailableBubblePieKeys();

    var events_plot = {};
    var timer = 0,
        highlightpoint,
        stacked = function(cType) {
            return chartObj.dataObject.isstacked(cType);
        };

    // custom points
    var CustomPoints_th;
    if (chartObj.cache.thresholdCustomPoints) {
        CustomPoints_th = new events.customPoints(chartObj);
        CustomPoints_th.addPoints(chartObj.cache.thresholdCustomPoints, true);
    }

    events_plot.initializeListener = function() {
        var eventHandler = chartObj.eventHandler;
        // let brushNZoomSupported = isFunction(chartObj.updateScale) || chartObj.isGeoMap;

        // let updateBrush = (getPropVal(systemConf.chart, 'brush.enabled') && brushNZoomSupported) || chartObj.plotBrush;
        // let updateZoom = (getPropVal(systemConf.chart, 'zoom.enabled') && brushNZoomSupported) || chartObj.plotZoom;

        tapOldInfo = chartObj.eventHandler.tapOldInfo = null; //Tap event in touch devices - to toggle (defaultMouseOverFunc & defaultMouseOutFunc) on tap;

        /**
         *
         * #ZC1831
         * Adding the g element as event element since d3 brush can need it.
         *
         * Previously we were calling the d3 zoom and brush first and then our chart events.
         * and we were adding it in brush-constructor and replace the  trackerElement variable later.
         *
         * But the problem is that d3 blocks all the intermediate events for touch device in zoom and brush.
         * That's why we are changing actual event element to a group element and bind the events accordingly.
         *
         */
        eventHandler.trackerGroup = createElement('g', '.trackerGroup', [1], wrap).attr('pointer-events', 'all');
        eventHandler.trackerElement = createElement('rect', '.eventhandler', [1], eventHandler.trackerGroup);

        eventHandler.tooltipElement = events.tooltip(tooltipele, systemConf, chartObj);
        eventHandler.tooltipElement.initialize();
        eventHandler.customPoints = CustomPoints_th;
        addClipToTracker();

        var trackerGroup = eventHandler.trackerGroup,
            trackerEle = eventHandler.trackerElement.style('fill-opacity', 0);

        var clickHandlers = {
                click: eventHandler.click,
                dblclick: eventHandler.doubleclick,
                contextmenu: eventHandler.contextMenu
            },
            moveHandlers = {
                mousemove: eventHandler.mouseMove,
                mouseout: eventHandler.mouseOut
            };

        attachevents(trackerGroup, clickHandlers, null, eventAttacherOptions, 'chart');
        attachevents(trackerGroup, moveHandlers, null, eventAttacherOptions, 'chartmove');
        attachevents(trackerGroup, moveHandlers, null, eventAttacherOptions);

        // if (updateBrush) {
        //     // set brush options
        //     setDefaultOptions('brush', chartObj.systemConf.chart, chartObj.globalProp);
        //     chartObj.plotBrush = events.plotBrush(chartObj);
        // } else {
        // }

        // touch events
        var touchHandlers = {
            tap: eventHandler.click,
            dbltap: eventHandler.doubleclick,
            singletouchmove: eventHandler.mouseMove,
            touchend: eventHandler.mouseOut
        };

        attachevents(trackerGroup, {}, touchHandlers, eventAttacherOptions);

        //
        // #ZC1831
        // enable geo zoom & brush
        // if (updateZoom) {
        //     // merge the default options for zoom.
        //     setDefaultOptions('zoom', chartObj.systemConf.chart, chartObj.globalProp);
        //     // set the brush first because the brush needs a entire separate group to work.
        //     // set the zoom
        //     chartObj.plotZoom = events.plotZoom(chartObj);
        // }

        // if (chartObj.scroll) {
        //     chartObj.scroll.render();
        // }

        // handle scroll
        var tooltip = chartObj.eventHandler.tooltipElement;
        var scrollEventName =
            'onwheel' in document.createElement('div') ? 'wheel' : document.onmousewheel !== undefined ? 'mousewheel' : 'DOMMouseScroll';
        var parentNode = chartObj.container.node(),
            lastPosition = null;
        chartObj.d3container.on(scrollEventName, function() {
            var currentPosition = parentNode.getBoundingClientRect();
            if (lastPosition) {
                if (lastPosition.left !== currentPosition.left || lastPosition.top !== currentPosition.top) {
                    tooltip.hide(tooltip.element);
                }
            }
            lastPosition = currentPosition;
        });

        // click able tooltip
        if (systemConf.tooltip.pointerEvents === 'auto') {
            var tooltipElement = eventHandler.tooltipElement.element;
            tooltipElement.on('mouseenter', function() {
                events_plot.mouseOnTooltip = true;
            });

            tooltipElement.on('mouseleave', function() {
                events_plot.mouseOnTooltip = false;
                eventHandler.mouseOut();
            });
        }
        events_plot.reverseLegendHighlight = new (Registry.getComponent('legend.ReverseLegendHighlight'))(chartObj);
        events_plot.unselectedHighlight = new UnSelectedHighlight(chartObj);

        if (chartObj.isGeoMap) {
            events_plot.mapEvents = new zmaps.mapEvents(chartObj);
        }
    };

    events_plot.updateParams = function(chartObj_new, isRedraw = true) {
        chartObj = chartObj_new;
        systemConf = chartObj.systemConf;
        plotarea = chartObj.plotarea;
        dataset = chartObj.dataset;
        xscale = chartObj.axes.x.eventscale;
        yscale = chartObj.axes.y.groups;
        barscale1 = chartObj.barscale1;
        barscale2 = chartObj.barscale2;
        color = chartObj.seriesColor;
        tooltipele = chartObj.tooltip.element;
        stacked = function(cType) {
            return chartObj.dataObject.isstacked(cType);
        };
        axisrotated = chartObj.axes.rotated;

        // ZC1829
        // if chart zoom is changed in redraw, initialize the events again. because the event element can change.
        // if ((getPropVal(systemConf.chart, 'zoom.enabled') === true && chartObj.isGeoMap) || chartObj.geoZoom || chartObj.geoBrush) {
        //     events_plot.initializeListener();
        // }

        wrap = chartObj.staticPaneSVG.selectAll('g.eventgroup');
        datatype = dataset.getDataType('x');

        tooltipEnabled = systemConf.tooltip.enabled === true;
        chartObj.eventHandler.tooltipElement = events.tooltip(chartObj.tooltip.element, systemConf, chartObj);
        chartObj.eventHandler.tooltipElement.initialize();
        chartObj.eventHandler.bulletEvents = null; //#ZC1009
        tooltipShared = tooltip_helpers.hasSharedSeries(chartObj); //#ZC998;
        events_plot.reverseLegendHighlight = new (Registry.getComponent('legend.ReverseLegendHighlight'))(chartObj);
        events_plot.unselectedHighlight = new UnSelectedHighlight(chartObj);

        // update custom points
        if (CustomPoints_th) {
            CustomPoints_th.removePoints(0, CustomPoints_th.points.length);
            CustomPoints_th.addPoints(chartObj.cache.thresholdCustomPoints, true);
        }

        if (chartObj.isAxisCategory || isPolarAxisCategory || chartObj.isGeoMap) {
            // drawBorderBox(chartObj.eventHandler.trackerElement, getPlotBoundingRect(chartObj));
            chartObj.positionEventTrackerElement();
            addClipToTracker();
        }

        if (chartObj.isGeoMap) {
            events_plot.mapEvents.initialize();
        }

        // render the scroll
        if (chartObj.scroll && chartObj_new.updating && isRedraw) {
            chartObj.scroll.render();
        }
    };

    events_plot.mouseMove = function(args) {
        var eventEle = args.element || this;
        if (chartObj.chartRendered) {
            var tooltip = chartObj.eventHandler.tooltipElement;
            if (tooltip != null && tooltip.timeout != null) {
                clearTimeout(tooltip.timeout); // To avoid flickering on  mouse move, clearing the timeout that makes the tooltip's display prop to none
            }
            // if (d3_event().defaultPrevented) return;
            if (chartObj.stopPropogate !== false) {
                d3_event().stopPropagation();
            }
            chartObj.eventHandler.targetEvent = d3_event();

            if (!chartObj.isAxisCategory && !isPolarAxisCategory && !chartObj.isGeoMap) {
                var node = chartObj.svg.node();
                chartObj.eventHandler.mousePos = getMousePosition(node, d3_event());
                var relMouseX = chartObj.eventHandler.mousePos[0];
                var relMouseY = chartObj.eventHandler.mousePos[1];

                var relmousePt = {
                    x: relMouseX,
                    y: relMouseY
                };

                var currentEle = d3_event().srcElement || d3_event().target;
                var elementID = d3_select(currentEle).attr('id');
                args = args.constructor === Object ? args : chartObj.eventHandler.oldRenderer;

                if (args) {
                    var seriesdata = chartObj.seriesdata;
                    var chartName = chartObj.chartTypes[0].name;

                    var plotOptions = seriesdata[args.seriesPos].plotoptions,
                        commonPlotOptions = systemConf.chart.plot.plotoptions[chartName],
                        eventType = chartObj.zcEventType || 'mousemove';
                    var mouseMoveFn = getEventConf(plotOptions, commonPlotOptions, chartObj, eventType);
                    var evt = d3_event() || chartObj.eventHandler.targetEvent;

                    if (defined(mouseMoveFn)) {
                        triggerUserCallback(mouseMoveFn, args, evt);
                    }

                    if (isDefaultAllowed(mouseMoveFn, evt)) {
                        chartObj.eventHandler.oldRenderer = args;
                        args.renderer.defaultMouseOverFunc(args.data, args.seriesPos, args.itemPos);
                    }

                    if (tooltipEnabled) {
                        var tooltipPos = {
                            left: relmousePt.x,
                            top: relmousePt.y
                        };
                        evt = getEventObject(d3_event());
                        var dLeft = relmousePt.x - evt.clientX,
                            dTop = relmousePt.y - evt.clientY;
                        tooltipPos.left -= dLeft;
                        tooltipPos.top -= dTop;
                        tooltipPos.dLeft = dLeft;
                        tooltipPos.dTop = dTop;
                        var tooltipArgs = {
                            seriesindex: args.seriesPos,
                            subseriespos: args.subseriesPos,
                            color: args.color || color,
                            svgFillFunc: args.renderer.fillColorFunc,
                            itempos: args.itemPos,
                            d: chartObj.dataObject.isHierarchy ? args.data.data || args.data : null
                        };
                        tooltip.show(chartObj, tooltipPos, null, true, tooltipArgs);
                    }
                }
            } else {
                if (chartObj.eventHandler.mousemovetimer) {
                    clearTimeout(chartObj.eventHandler.mousemovetimer);
                }
                var node = chartObj.eventHandler.trackerElement.node();
                chartObj.eventHandler.mousePos = getMousePosition(node, d3_event());
                chartObj.eventHandler.mousemovetimer = setTimeout(function() {
                    if (chartObj.isGeoMap) {
                        events_plot.mapEvents.handleEvent(chartObj.eventHandler.mousePos, 'mousemove', chartObj.eventHandler.targetEvent);
                    } else {
                        var containsBarchart = chartObj.dataObject.noofBarEnabledSeries > 0;
                        var func =
                            datatype === ORDINAL_DATATYPE && (containsBarchart || tooltipShared) ? ordinalDatamouseMove : xydatamouseMove;
                        func(chartObj.eventHandler.mousePos, chartObj.eventHandler.targetEvent, node);
                    }
                }, timer);
            }
        }

        // reverse hovering
        var currentPoint =
            chartObj.isAxisCategory || isPolarAxisCategory || chartObj.isGeoMap
                ? chartObj.eventHandler.oldInfo && chartObj.eventHandler.oldInfo.nearestPoint
                : args;
        events_plot.reverseLegendHighlight.highlight(currentPoint);
        events_plot.unselectedHighlight.highlight(currentPoint);
    };

    events_plot.mouseOut = function(renderer, data, index, showTooltip, forceMouseOut) {
        if (events_plot.mouseOnTooltip === true) {
            return;
        }

        var seriesdata = chartObj.seriesdata,
            eventEle = this;
        index = index || 0; // Touch double tap
        var chartName = chartObj.chartTypes[0].name;
        var plotOptions =
                chartObj.isNonAxisMultiSeriesCategory && !isPolarAxisCategory ? seriesdata[index].plotoptions : seriesdata[0].plotoptions,
            commonPlotOptions = systemConf.chart.plot.plotoptions[chartName];
        var oldInfo = chartObj.eventHandler.oldInfo,
            tooltipHideDelay = systemConf.tooltip.hideDelay, //document-events.js also uses this
            closestPoint = {};

        var event = d3_event() || chartObj.eventHandler.targetEvent,
            tooltip = chartObj.eventHandler.tooltipElement;

        events_plot.unselectedHighlight.reset();
        // for geo map
        if (chartObj.isGeoMap) {
            return events_plot.mapEvents.handleEvent(null, 'mouseout', event);
        }

        // check custom point
        if (oldInfo) {
            closestPoint = pick(oldInfo.nearestPoint, closestPoint);
            if (oldInfo.customPoint) {
                commonPlotOptions = oldInfo.customPoint;
            }
        }

        var mouseOutFn = getEventConf(plotOptions, commonPlotOptions, chartObj, 'mouseout');

        if (forceMouseOut || chartObj.chartRendered) {
            var withIn = 0;

            if (defined(mouseOutFn)) {
                if (!chartObj.isAxisCategory || (oldInfo != null && !oldInfo.isNoInfo)) {
                    var point;
                    if (oldInfo && defined(oldInfo.nearestPoint)) {
                        point = oldInfo.nearestPoint;
                    } else if (data) {
                        point = { seriesPos: 0, itemPos: index, point: data.data };
                    }

                    if (defined(point)) {
                        triggerUserCallback(mouseOutFn, point, event);
                    }
                } else if (oldInfo && defined(oldInfo.customPoint)) {
                    triggerUserCallback(mouseOutFn, oldInfo.customPoint, event);
                }
            }

            if (isDefaultAllowed(mouseOutFn, event)) {
                if (chartObj.isAxisCategory || isPolarAxisCategory) {
                    event = d3_event();

                    var node = chartObj.eventHandler.trackerElement.node();
                    var mouse = event != null ? getMousePosition(node, event) : chartObj.eventHandler.mousePos;
                    var relativeelement = bbox(node);
                    var relMouseX = axisrotated ? mouse[1] : mouse[0];
                    var relMouseY = axisrotated ? mouse[0] : mouse[1];
                    var relmousePt = {
                        x: relMouseX,
                        y: relMouseY
                    };

                    withIn = isWithin(relativeelement, relmousePt);
                }

                var relatedTarget = event != null ? event.relatedTarget : null;
                var target = event != null ? event.target : null;

                if (
                    chartObj.isAxisCategory &&
                    withIn == 0 &&
                    relatedTarget != null &&
                    tooltip != null &&
                    (target == tooltip.element.node() ||
                        target == tooltip.element.selectAll('div#tooltipcontent').node() ||
                        relatedTarget == tooltip.element.node() ||
                        relatedTarget == tooltip.element.selectAll('div#tooltipcontent').node())
                ) {
                    return;
                } else {
                    if (!chartObj.isAxisCategory && !isPolarAxisCategory) {
                        renderer = renderer || chartObj.renderer[index];

                        if (
                            eventEle.id == 'zctooltip' ||
                            (relatedTarget != null &&
                                tooltip != null &&
                                (target == tooltip.element.node() ||
                                    target == tooltip.element.selectAll('div#tooltipcontent').node() ||
                                    target == tooltip.element.selectAll('div#zctooltipafter').node() ||
                                    target == tooltip.element.selectAll('div#zctooltipbefore').node() ||
                                    relatedTarget == tooltip.element.node() ||
                                    relatedTarget == tooltip.element.selectAll('div#tooltipcontent').node() ||
                                    relatedTarget == tooltip.element.selectAll('div#zctooltipafter').node() ||
                                    relatedTarget == tooltip.element.selectAll('div#zctooltipbefore').node()))
                        ) {
                            var actualRenderer = chartObj.eventHandler.oldRenderer;
                            if (actualRenderer != null) {
                                actualRenderer.renderer.defaultMouseOutFunc(
                                    actualRenderer.data,
                                    chartObj.isNonAxisMultiSeriesCategory ? actualRenderer.seriesPos : actualRenderer.itemPos,
                                    forceMouseOut
                                );
                            }
                        } else {
                            renderer.defaultMouseOutFunc(data, index, forceMouseOut);
                        }
                    }
                }
                // // mouse out callback
                // if (mouseOutFn != null) {
                //     if ((!chartObj.isAxisCategory) || (oldInfo != null && !oldInfo.isNoInfo)) {
                //         var evt = event || chartObj.eventHandler.targetEvent,
                //             dt = (data) ? {
                //                     "data": chartObj.dataset.getMetaInfoObj(chartObj, (data.data || data), index)
                //                 } : null;
                //         invokeFunction(mouseOutFn, evt, dt, chartObj);
                //     }

                // }
            }

            // always hide tooltip on mouse out
            // because we always show the tooltip on mouse enter.

            clearTimeout(chartObj.eventHandler.mouseouttimer);
            chartObj.eventHandler.mouseouttimer = setTimeout(function() {
                if (showTooltip !== true) {
                    clearTimeout(chartObj.eventHandler.tooltipElement.timeout);
                    chartObj.eventHandler.tooltipElement.timeout = setTimeout(function() {
                        hideHighlighterPoints();
                        tooltip.hide();
                    }, pick(tooltipHideDelay, 150));
                }
            }, timer);

            events_plot.reverseLegendHighlight.reset();
            events_plot.hideCursor(chartObj.eventHandler.trackerElement);
        }
    };

    events_plot.click = function(renderer, data, seriesPos, itemPos, args) {
        if (chartObj.chartRendered) {
            var seriesdata = chartObj.seriesdata;
            var isTouchEvent = d3_event() && d3_event().touches,
                eventType = chartObj.zcEventType || 'click';

            if (isTouchEvent && eventType !== 'taphold') {
                documentTapEvent(chartObj);

                $ZCG.clickedTime = new Date().getTime();
                $ZCG.clickedElement = this;
                $ZCG.clickArguments = [renderer, data, seriesPos, itemPos];
                $ZCG.previousChartObj = chartObj;
            }

            var tooltip = chartObj.eventHandler.tooltipElement,
                evt = d3_event() || chartObj.eventHandler.targetEvent;

            if (!chartObj.isAxisCategory && !isPolarAxisCategory && !chartObj.isGeoMap) {
                var chartName = chartObj.chartTypes[0].name;
                var plotOptions = seriesdata[chartObj.isNonAxisMultiSeriesCategory ? seriesPos : 0].plotoptions,
                    commonPlotOptions = systemConf.chart.plot.plotoptions[chartName];

                var clickFn = getEventConf(plotOptions, commonPlotOptions, chartObj, eventType);

                if (defined(clickFn)) {
                    triggerUserCallback(clickFn, args, evt);
                }

                // hide tooltip if default is not prevnted
                if (isDefaultAllowed(clickFn, evt)) {
                    tooltip.hide(tooltip.element);
                    events_plot.hideCursor(chartObj.eventHandler.trackerElement);
                }

                if (isTouchEvent && eventType !== 'taphold' && isDefaultAllowed(clickFn, evt)) {
                    tapOldInfo = chartObj.eventHandler.tapOldInfo;
                    if (
                        !tapOldInfo ||
                        (args.level
                            ? args.level != tapOldInfo.level
                            : tapOldInfo.itemPos != itemPos || tapOldInfo.seriesPos != seriesPos || tapOldInfo.subseriesindex != 0)
                    ) {
                        renderer.defaultMouseOverFunc(data, seriesPos, itemPos);

                        if (tooltipEnabled) {
                            var node = chartObj.svg.node();
                            chartObj.eventHandler.mousePos = getMousePosition(node, d3_event());
                            var relMouseX = chartObj.eventHandler.mousePos[0];
                            var relMouseY = chartObj.eventHandler.mousePos[1];

                            var relmousePt = {
                                x: relMouseX,
                                y: relMouseY
                            };

                            var tooltipPos = {
                                left: relmousePt.x,
                                top: relmousePt.y
                            };
                            var evt = getEventObject(d3_event());
                            var dLeft = relmousePt.x - evt.clientX,
                                dTop = relmousePt.y - evt.clientY;
                            tooltipPos.left -= dLeft;
                            tooltipPos.top -= dTop;
                            tooltipPos.dLeft = dLeft;
                            tooltipPos.dTop = dTop;
                            var tooltipArgs = {
                                seriesindex: args.seriesPos,
                                subseriespos: args.subseriesPos,
                                color: args.color || color,
                                svgFillFunc: args.renderer.fillColorFunc,
                                itempos: args.itemPos,
                                d: chartObj.dataObject.isHierarchy ? args.data.data || args.data : null
                            };
                            tooltip.show(chartObj, tooltipPos, null, true, tooltipArgs);
                        }

                        chartObj.eventHandler.tapOldInfo = {
                            seriesPos: seriesPos,
                            subseriesindex: 0,
                            itemPos: itemPos,
                            level: args.level
                        };
                    } else {
                        renderer.defaultMouseOutFunc(data, itemPos);
                        chartObj.eventHandler.tapOldInfo = null;
                    }
                }
            } else {
                var event = d3_event();
                element = chartObj.eventHandler.trackerElement.node();
                var mouse = getMousePosition(element, event) || chartObj.eventHandler.mousePos;

                // geo map things
                if (chartObj.isGeoMap) {
                    return events_plot.mapEvents.handleEvent(mouse, 'click', event);
                }

                var relativeelement = bbox(element);
                var relMouseX = mouse[0] - relativeelement.x - (isPolarAxisCategory ? chartObj.plotarea_approx.x : 0);
                var relMouseY = mouse[1] - relativeelement.y - (isPolarAxisCategory ? chartObj.plotarea_approx.y : 0);
                var relmousePt = {
                    x: relMouseX,
                    y: relMouseY
                };

                element = this;
                var containsBarchart = chartObj.dataObject.noofBarEnabledSeries > 0;
                var func = datatype === ORDINAL_DATATYPE && containsBarchart ? getOrdinalDataPoint : getClosestPoint;
                var nearestPoint = func(relmousePt, xscale, yscale, systemConf);

                var seriespos = nearestPoint.seriesPos;
                var itemPos = nearestPoint.itemPos;
                var subseriesindex = isArray(nearestPoint.subseriesindex)
                    ? nearestPoint.subseriesindex[seriespos]
                    : nearestPoint.subseriesindex;
                var customPoint = pick(nearestPoint.customPoint, nearestPoint.currentQualitativeRangePoint);

                // TODO: Need to in common way for both series point and custom point. Priority: HIGH
                if (defined(customPoint)) {
                    var thresholdClick = getEventConf(null, customPoint, chartObj, eventType);
                    if (defined(thresholdClick)) {
                        triggerUserCallback(thresholdClick, customPoint, event);
                    }
                    chartObj.eventHandler.oldInfo = { customPoint: customPoint, isNoInfo: true };
                }

                if (seriespos == null && !defined(customPoint)) {
                    if (isTouchEvent) {
                        chartObj.eventHandler.mouseOut();
                        chartObj.eventHandler.tapOldInfo = null;
                    }
                    return;
                }

                if (defined(seriespos)) {
                    var type = chartObj.chartTypes[seriespos].value;
                    var plotOptions = seriesdata[seriespos].plotoptions,
                        commonPlotOptions = systemConf.chart.plot.plotoptions[charttypenames.get(type)];

                    var clickFn = getEventConf(
                            plotOptions,
                            commonPlotOptions,
                            chartObj,
                            eventType,
                            nearestPoint.currentQualitativeRangePoint
                        ),
                        cursor = getEventConf(plotOptions, commonPlotOptions, chartObj, 'cursor');
                }

                // #ZC1115 --> #ZC1127
                if (defined(clickFn)) {
                    var yscindex = getParsedYAxisOrder(seriesdata[seriespos].yaxiscolumnorder),
                        ysc = yscale[yscindex].eventscale,
                        data = getCurrentDataPoint(seriespos, subseriesindex, itemPos, nearestPoint.itemPosShared);

                    if (isBarChartType(type)) {
                        var highlightPoint = nearestPoint.highlightpoint || getPointRectangle(seriespos, subseriesindex, itemPos);
                        events_plot.addCursor(cursor);
                        triggerUserCallback(clickFn, nearestPoint, evt, {
                            x: highlightPoint.x,
                            y: highlightPoint.y
                        });
                    } else {
                        var xval = chartObj.dataObject.getXYvalue(axisrotated, stacked(type), xscale, ysc, yscindex, data, 'x');
                        var yval = chartObj.dataObject.getXYvalue(axisrotated, stacked(type), xscale, ysc, yscindex, data, 'y');
                        var mindist =
                            isAreaChartType(type) || isPolarAxisCategory
                                ? null
                                : lineDistance(
                                      {
                                          x: xval,
                                          y: yval
                                      },
                                      relmousePt
                                  );
                        var rad = isBubbleChartType(type) ? chartObj.zscale(chartObj.dataset.getZ(data)) / 2 : 10;
                        if (mindist <= rad) {
                            events_plot.addCursor(cursor);
                            //						var extraParams = "GC 0 "+tooltip.formatValue(systemConf,xdata,"x")+",2 "+seriesdata[seriespos].data[subseriesindex][itemPos][dataset.indices.clr]+",";

                            triggerUserCallback(clickFn, nearestPoint, evt, {
                                x: xval,
                                y: yval
                            });
                        } else {
                            events_plot.hideCursor(chartObj.eventHandler.trackerElement);
                            return;
                        }
                    }
                }

                // hide tooltip if default is not prevnted
                if (isDefaultAllowed(clickFn, evt)) {
                    tooltip.hide(tooltip.element);
                    events_plot.hideCursor(chartObj.eventHandler.trackerElement);
                }

                if (isTouchEvent && eventType !== 'taphold' && isDefaultAllowed(clickFn, evt)) {
                    tapOldInfo = chartObj.eventHandler.tapOldInfo;
                    var currentQualitativeRangePoint = nearestPoint.currentQualitativeRangePoint || {};
                    var currentQualitativeRangePoint_old = (tapOldInfo && tapOldInfo.currentQualitativeRangePoint) || {};

                    var handleMouseMove =
                        !tapOldInfo ||
                        tapOldInfo.itemPos != itemPos ||
                        tapOldInfo.seriesPos != seriespos ||
                        tapOldInfo.subseriesindex != subseriesindex ||
                        currentQualitativeRangePoint.rangeIndex != currentQualitativeRangePoint_old.rangeIndex ||
                        currentQualitativeRangePoint.targetIndex != currentQualitativeRangePoint_old.targetIndex ||
                        ((nearestPoint.highlightpoint && nearestPoint.highlightpoint.type) || currentQualitativeRangePoint.type) !=
                            currentQualitativeRangePoint_old.type;
                    if (handleMouseMove) {
                        chartObj.eventHandler.mouseMove({});

                        chartObj.eventHandler.tapOldInfo = {
                            seriesPos: seriespos,
                            subseriesindex: subseriesindex,
                            itemPos: itemPos,
                            currentQualitativeRangePoint: currentQualitativeRangePoint,
                            type: (nearestPoint.highlightpoint && nearestPoint.highlightpoint.type) || currentQualitativeRangePoint.type
                        };
                    } else {
                        chartObj.eventHandler.mouseOut();
                        chartObj.eventHandler.tapOldInfo = null;
                    }
                }
            }
        }
    };

    events_plot.doubleclick = function(data, index, args) {
        if (chartObj.chartRendered) {
            if (chartObj.noPropogate) {
                d3_event().stopPropagation();
            }

            var isTouchEvent = d3_event() && d3_event().touches;
            if (isTouchEvent && !chartObj.isGeoMap) {
                chartObj.eventHandler.mouseOut();
                chartObj.eventHandler.tapOldInfo = null;
            }

            var evt = d3_event() || chartObj.eventHandler.targetEvent;
            var tooltip = chartObj.eventHandler.tooltipElement;

            var seriesdata = chartObj.seriesdata;
            var eventType = chartObj.zcEventType || 'doubleclick';
            if (!chartObj.isAxisCategory && !isPolarAxisCategory && !chartObj.isGeoMap) {
                var chartName = chartObj.chartTypes[0].name;
                var plotOptions = seriesdata[chartObj.isNonAxisMultiSeriesCategory ? index : 0].plotoptions,
                    commonPlotOptions = systemConf.chart.plot.plotoptions[chartName];
                var dblClickFn = getEventConf(plotOptions, commonPlotOptions, chartObj, eventType);

                if (defined(dblClickFn)) {
                    triggerUserCallback(dblClickFn, { seriesPos: 0, itemPos: index, point: data.data }, evt);
                }
                if (isDefaultAllowed(dblClickFn, evt)) {
                    addNoteFunction(args);
                }
            } else {
                element = chartObj.eventHandler.trackerElement.node();
                var mouse = getMousePosition(element, d3_event());

                // geo map things
                if (chartObj.isGeoMap) {
                    return events_plot.mapEvents.handleEvent(mouse, 'doubleclick', event);
                }

                var relativeelement = bbox(element);
                var relMouseX = mouse[0] - relativeelement.x - (isPolarAxisCategory ? chartObj.plotarea_approx.x : 0);
                var relMouseY = mouse[1] - relativeelement.y - (isPolarAxisCategory ? chartObj.plotarea_approx.y : 0);
                var relmousePt = {
                    x: relMouseX,
                    y: relMouseY
                };

                element = this;
                var containsBarchart = chartObj.dataObject.noofBarEnabledSeries > 0;
                var func = datatype === ORDINAL_DATATYPE && containsBarchart ? getOrdinalDataPoint : getClosestPoint;
                var nearestPoint = func(relmousePt, xscale, yscale, systemConf),
                    customPoint = pick(nearestPoint.customPoint, nearestPoint.currentQualitativeRangePoint);

                // TODO: Need to in common way for both series point and custom point. Priority: HIGH
                if (defined(customPoint) && !defined(seriespos)) {
                    var thresholdClick = getEventConf(null, customPoint, chartObj, eventType);
                    if (defined(thresholdClick)) {
                        triggerUserCallback(thresholdClick, customPoint, evt);
                    }
                    chartObj.eventHandler.oldInfo = { customPoint: customPoint, isNoInfo: true };
                    return;
                }

                var seriespos = nearestPoint.seriesPos;
                if (seriespos == null) {
                    return;
                }
                var itemPos = nearestPoint.itemPos,
                    subseriesindex =
                        nearestPoint.subseriesindex instanceof Array ? nearestPoint.subseriesindex[seriespos] : nearestPoint.subseriesindex,
                    data = getCurrentDataPoint(seriespos, subseriesindex, itemPos, nearestPoint.itemPosShared),
                    type = chartObj.chartTypes[seriespos].value,
                    plotOptions = seriesdata[seriespos].plotoptions,
                    commonPlotOptions = systemConf.chart.plot.plotoptions[charttypenames.get(type)],
                    dblClickFn = getEventConf(
                        plotOptions,
                        commonPlotOptions,
                        chartObj,
                        eventType,
                        nearestPoint.currentQualitativeRangePoint
                    ),
                    cursor = getEventConf(plotOptions, commonPlotOptions, chartObj, 'cursor');

                if (isBarChartType(type)) {
                    events_plot.addCursor(cursor);
                    if (defined(dblClickFn)) {
                        triggerUserCallback(dblClickFn, nearestPoint, evt);
                    }

                    if (isDefaultAllowed(dblClickFn, evt)) {
                        var highlightPoint = pick(nearestPoint.highlightpoint, {}),
                            currentSeriesdata = getCurrentDataPoint(seriespos, subseriesindex, itemPos, nearestPoint.itemPosShared, true);
                        addNoteFunction(null, seriespos, itemPos, currentSeriesdata, highlightPoint.x, highlightPoint.y);
                    }
                } else {
                    var yscindex = getParsedYAxisOrder(seriesdata[seriespos].yaxiscolumnorder);
                    var ysc = yscale[yscindex].eventscale;
                    var xval =
                        chartObj.dataObject.getXYvalue(axisrotated, stacked(type), xscale, ysc, yscindex, data, 'x') +
                        (isPolarAxisCategory ? chartObj.plotarea_approx.x : 0);
                    var yval =
                        chartObj.dataObject.getXYvalue(axisrotated, stacked(type), xscale, ysc, yscindex, data, 'y') +
                        (isPolarAxisCategory ? chartObj.plotarea_approx.y : 0);
                    var mindist =
                        isAreaChartType(type) || isPolarAxisCategory
                            ? null
                            : lineDistance(
                                  {
                                      x: xval,
                                      y: yval
                                  },
                                  relmousePt
                              );
                    var rad = isBubbleChartType(type) ? chartObj.zscale(chartObj.dataset.getZ(data)) / 2 : 10;
                    if (mindist <= rad) {
                        events_plot.addCursor(cursor);

                        if (defined(dblClickFn)) {
                            triggerUserCallback(dblClickFn, nearestPoint, evt, { x: xval, y: yval });
                        }

                        if (isDefaultAllowed(dblClickFn, evt)) {
                            var currentSeriesdata = getCurrentDataPoint(
                                seriespos,
                                subseriesindex,
                                itemPos,
                                nearestPoint.itemPosShared,
                                true
                            );
                            addNoteFunction(null, seriespos, itemPos, currentSeriesdata, xval, yval);
                        }
                    }

                    if (isDefaultAllowed(dblClickFn, evt)) {
                        events_plot.hideCursor(chartObj.eventHandler.trackerElement);
                    }
                }
            }

            if (isDefaultAllowed(dblClickFn, evt)) {
                tooltip.hide(tooltip.element);
                events_plot.hideCursor(chartObj.eventHandler.trackerElement);
            }
        }
    };

    events_plot.contextMenu = function(args) {
        var contextMenuFn = systemConf.chart.plot.events.contextmenu;
        if (contextMenuFn != null) {
            var evt = d3_event() || chartObj.eventHandler.targetEvent;
            invokeFunction(contextMenuFn, evt, chartObj);
            d3_event().preventDefault();
        }
    };

    events_plot.addCursor = function(cur) {
        var trackerElement = getEventTrackerElement();
        cur = cur || (chartObj.plotBrush ? 'crosshair' : POINTER);
        trackerElement.style('cursor', cur);
    };

    events_plot.hideCursor = function() {
        var trackerElement = getEventTrackerElement();
        trackerElement.style('cursor', 'default');
    };

    function getEventTrackerElement() {
        return chartObj.isAxisCategory && (chartObj.systemConf.chart.width || chartObj.systemConf.chart.height)
            ? defaultScrollElement
            : chartObj.eventHandler.trackerElement;
    }

    function addClipToTracker() {
        if (isPolarAxisCategory) {
            var plotarea_approx = chartObj.plotarea_approx,
                plotarea = chartObj.plotarea,
                properties = {
                    name: 'polarClip',
                    id: chartObj.id,
                    cx: plotarea_approx.x + plotarea.left,
                    cy: plotarea_approx.y + plotarea.top,
                    r: plotarea_approx.radius
                },
                args = {
                    chartObj: chartObj
                },
                clipID = effects.get(effects.svg_clips, CIRCLE_ELEMENT, properties, args);

            chartObj.eventHandler.trackerElement.attr('clip-path', `url(${chartObj.defsLocation}#${clipID})`);

            if (!window.PointerEvent && chartObj.plot.renderer.mode === 'canvas') {
                //ZC1533 pointer-events: none is not working in ie9, ie10
                pointerEventsNonePolyfill(
                    chartObj.container.selectAll('canvas.zc-labelholder-canvas,canvas.highlight-canvas,canvas.zc-canvas')
                );
            }
        }
    }

    function ordinalDatamouseMove(mouse, event, element) {
        var relativeelement = bbox(element);
        var relMouseX = mouse[0] - relativeelement.x;
        var relMouseY = mouse[1] - relativeelement.y;

        var relmousePt = {
            x: relMouseX,
            y: relMouseY
        };
        var cursorFollowDist = 10;
        var nearestPoint = getOrdinalDataPoint(relmousePt, xscale, yscale, systemConf, true);
        var seriesdata = chartObj.seriesdata;
        var seriespos = nearestPoint.seriesPos,
            xpos = nearestPoint.itemPos,
            subseriesindex = nearestPoint.subseriesindex[seriespos],
            subseriesLength = nearestPoint.subseriesLength[seriespos],
            xpos_shared = nearestPoint.itemPosShared,
            currentQualitativeRangePoint = nearestPoint.currentQualitativeRangePoint;

        highlightpoint = nearestPoint.highlightpoint;
        levelMarkerPoint = nearestPoint.levelMarkerPoint;
        var oldInfo = chartObj.eventHandler.oldInfo;

        var tooltip = chartObj.eventHandler.tooltipElement,
            customPoint = nearestPoint.customPoint;
        event = getEventObject(event);

        // handle custom point
        if (defined(customPoint) && CustomPoints_th) {
            if (customPoint.type === 'threshold-marker') {
                var thresholdMouseMove = getEventConf(null, customPoint, chartObj, 'mousemove');
                if (defined(thresholdMouseMove)) {
                    triggerUserCallback(thresholdMouseMove, customPoint, event);
                }
                CustomPoints_th.hoverPoint(customPoint, event, relmousePt);
            }
            chartObj.eventHandler.oldInfo = { customPoint: customPoint, isNoInfo: true };
            return;
        } else if (defined(currentQualitativeRangePoint) && systemConf.tooltip.enabled !== false) {
            // TODO NEED TO MERGE WITH CUSTOM POINT
            if (currentQualitativeRangePoint.type === 'marker' || !defined(seriespos)) {
                var tooltipPosition = getTooltipPosition(
                        event,
                        relmousePt,
                        0,
                        0,
                        chartObj,
                        nearestPoint,
                        axisrotated,
                        tooltipShared,
                        {},
                        true,
                        false,
                        false
                    ),
                    text = events_plot.bulletEvents.getTooltipContent(currentQualitativeRangePoint, 0);

                var bulletEvent = getEventConf(null, currentQualitativeRangePoint, chartObj, 'mousemove');
                if (defined(bulletEvent)) {
                    triggerUserCallback(bulletEvent, currentQualitativeRangePoint, event);
                }

                hideHighlighterPoints();
                tooltipPosition.helpers.inheritColors = true;
                tooltipPosition.helpers.itempos = currentQualitativeRangePoint.itemIndex;
                tooltipPosition.helpers.chartType = currentQualitativeRangePoint.chartType;
                tooltip.show(chartObj, tooltipPosition, text, false, tooltipPosition.helpers);
                chartObj.eventHandler.oldInfo = { customPoint: currentQualitativeRangePoint, isNoInfo: true };
                return;
            }
        }

        if (seriespos == null) {
            chartObj.eventHandler.mouseOut();
            chartObj.eventHandler.oldInfo = { isNoInfo: true };
            return;
        }

        var type = chartObj.chartTypes[seriespos].value,
            isRangeChart = type == charttype.arearange,
            isBarType = isBarChartType(type),
            isAreaType = isAreaChartType(type),
            isBubbleType = isBubbleChartType(type),
            yrangePos = isRangeChart ? 1 : null,
            yscindex = getParsedYAxisOrder(seriesdata[seriespos].yaxiscolumnorder),
            ysc = yscale[yscindex].eventscale,
            data = getCurrentDataPoint(seriespos, subseriesindex, xpos, xpos_shared),
            xdata = tooltipShared ? xscale.domain()[xpos] : dataset.getX(data), //chartObj.dataObject.getCategories("x")[xpos];
            ydata = dataset.getY(data, yscindex),
            xval = chartObj.dataObject.getXYvalue(axisrotated, stacked(type), xscale, ysc, yscindex, data, 'x', null, yrangePos),
            yval = chartObj.dataObject.getXYvalue(axisrotated, stacked(type), xscale, ysc, yscindex, data, 'y', null, yrangePos);

        if (!isBarType && isWithinPlotarea(xval, yval) == null) {
            return;
        }
        // zcareahighlight

        if (oldInfo != null && !oldInfo.isNoInfo) {
            var oldNearestPoint = oldInfo.nearestPoint;
            var oldType = getChartType(oldInfo.type || chartObj.userdata.seriesdata.type, NUMERIC_DATATYPE);
            if (
                oldNearestPoint &&
                oldNearestPoint.seriesPos == nearestPoint.seriesPos &&
                oldNearestPoint.itemPos == nearestPoint.itemPos &&
                oldNearestPoint.subseriesindex == nearestPoint.subseriesindex &&
                oldType == type
            ) {
                wrap.selectAll('#zcbarhighlight,#zcareahighlight').style('display', '');
                tooltipele.style('visibility', tooltipEnabled ? 'visible' : 'hidden');
                if (crosshair_enabled(chartObj)) {
                    chartObj.staticPaneSVG.selectAll('g.crosshair').style('visibility', 'visible');
                }
                if (cursorFollowDist > distance) {
                    events_plot.addCursor(oldInfo.cursor);
                } else {
                    events_plot.hideCursor(chartObj.eventHandler.trackerElement);
                }
                return;
            }
        }

        var distance =
            isBarType || isAreaType || isBubbleType || chartObj.dataObject.isPolarAxisCategory
                ? 0
                : lineDistance(
                      {
                          x: xval,
                          y: yval
                      },
                      relmousePt
                  );
        var seriesLen = tooltipShared ? chartObj.seriesdata.length : 1;
        //highlighting new Info
        wrap.selectAll('#zcscatterhighlight')
            .selectAll('*')
            .remove();
        wrap.selectAll('#zcbarhighlight')
            .selectAll('*')
            .remove();
        wrap.selectAll('#zcareahighlight')
            .selectAll('*')
            .remove();
        var tooltipSharedInfo = d3_map();

        for (var i = 0; i < seriesLen; i++) {
            var curSeriespos = tooltipShared ? i : seriespos;
            if (tooltipShared && subseriesLength > 1) {
                for (var j = 0; j < subseriesLength; j++) {
                    hoverSeries(
                        chartObj,
                        curSeriespos,
                        xdata,
                        xpos,
                        j,
                        nearestPoint,
                        tooltipShared,
                        cursorFollowDist > distance,
                        tooltipSharedInfo
                    );
                }
            } else {
                hoverSeries(
                    chartObj,
                    curSeriespos,
                    xdata,
                    xpos,
                    subseriesindex,
                    nearestPoint,
                    tooltipShared,
                    cursorFollowDist > distance,
                    tooltipSharedInfo
                );
            }
        }

        // Tooltip shown values
        if (tooltipEnabled) {
            var tooltipPos = getTooltipPosition(
                event,
                relmousePt,
                xval,
                yval,
                chartObj,
                nearestPoint,
                axisrotated,
                tooltipShared,
                tooltipSharedInfo,
                true,
                false,
                isMultiPieChartType(chartObj.chartTypes[seriespos].value) &&
                    chartObj.subrenderer[subseriesindex][bubblepieUtils.getKey(data, seriespos, chartObj)].fillColorFunc
            );
            tooltip.show(chartObj, tooltipPos, null, true, tooltipPos.helpers);
        }
    }

    function xydatamouseMove(mouse, event, element) {
        var cursorFollowDist = 10;
        var relativeelement = bbox(element);
        var relMouseX = mouse[0] - relativeelement.x - (isPolarAxisCategory ? chartObj.plotarea_approx.x : 0);
        var relMouseY = mouse[1] - relativeelement.y - (isPolarAxisCategory ? chartObj.plotarea_approx.y : 0);
        var relmousePt = {
            x: relMouseX,
            y: relMouseY
        };

        var oldInfo = chartObj.eventHandler.oldInfo;

        var evt = getEventObject(event),
            cur_container_top = relmousePt.y - event.clientY,
            cur_container_left = relmousePt.x - event.clientX;

        var nearestPoint = tooltipShared
            ? bisectX(relmousePt, xscale, yscale, systemConf)
            : getClosestPoint(relmousePt, xscale, yscale, systemConf);
        // var bisect = bisectX(relmousePt,xscale,yscale,systemConf)

        var seriespos = nearestPoint.seriesPos;
        var itemPos = nearestPoint.itemPos;
        var subseriesindex = nearestPoint.subseriesindex;
        var tooltip = chartObj.eventHandler.tooltipElement,
            customPoint = nearestPoint.customPoint;

        // handle custom point
        if (defined(customPoint) && CustomPoints_th) {
            if (customPoint.type === 'threshold-marker') {
                var thresholdMouseMove = getEventConf(null, customPoint, chartObj, 'mousemove');
                if (defined(thresholdMouseMove)) {
                    triggerUserCallback(thresholdMouseMove, customPoint, event);
                }
                CustomPoints_th.hoverPoint(customPoint, event, relmousePt);
            }
            chartObj.eventHandler.oldInfo = { customPoint: customPoint, isNoInfo: true };
            return;
        }

        if (!defined(seriespos)) {
            chartObj.eventHandler.mouseOut();
            chartObj.eventHandler.oldInfo = { isNoInfo: true };
            return;
        }

        var seriesdata = chartObj.seriesdata;
        var yscindex = getParsedYAxisOrder(seriesdata[seriespos].yaxiscolumnorder);
        var ysc = yscale[yscindex].scale;
        var type = chartObj.chartTypes[seriespos].value;
        var isRangeChart = type == charttype.arearange;
        var isAreaType = isAreaChartType(type);
        var isBubbleType = isBubbleChartType(type);
        var yrangePos = isRangeChart ? 1 : null;
        var rangeBand = xscale.bandwidth != null ? xscale.bandwidth() : null;
        var renderer = chartObj.renderer[seriespos];
        var currentSeriesdata =
            renderer.seriesdataAfterUpdate != null
                ? renderer.seriesdataAfterUpdate[subseriesindex]
                : seriesdata[seriespos].data[subseriesindex];
        var data = currentSeriesdata[itemPos];
        var xdata = chartObj.dataset.getX(data);

        var xval = chartObj.dataObject.getXYvalue(axisrotated, stacked(type), xscale, ysc, yscindex, data, 'x', null, yrangePos);
        var yval = chartObj.dataObject.getXYvalue(axisrotated, stacked(type), xscale, ysc, yscindex, data, 'y', null, yrangePos);

        var isAreaChart;
        if (!isPolarAxisCategory && isWithinPlotarea(xval, yval) == null) {
            return;
        }

        var distance =
            isAreaType || isBubbleType || chartObj.dataObject.isPolarAxisCategory
                ? 0
                : lineDistance(
                      {
                          x: xval,
                          y: yval
                      },
                      relmousePt
                  );
        // retaining oldinfo
        if (oldInfo != null && !tooltipShared) {
            var oldNearestPoint = oldInfo.nearestPoint;
            var oldType = getChartType(oldInfo.type || chartObj.userdata.seriesdata.type, NUMERIC_DATATYPE);
            if (
                oldNearestPoint &&
                oldNearestPoint.seriesPos == nearestPoint.seriesPos &&
                oldNearestPoint.itemPos == nearestPoint.itemPos &&
                oldNearestPoint.subseriesindex[oldNearestPoint.seriesPos] == nearestPoint.subseriesindex[nearestPoint.seriesPos] &&
                oldType == type &&
                oldInfo.containerLeft === cur_container_left &&
                oldInfo.containerTop === cur_container_top
            ) {
                wrap.selectAll(isAreaType ? '#zcscatterhighlight,#zcareahighlight' : '#zcscatterhighlight').style('display', '');
                tooltipele.style('visibility', tooltipEnabled ? 'visible' : 'hidden');
                if (crosshair_enabled(chartObj)) {
                    chartObj.staticPaneSVG.selectAll('g.crosshair').style('visibility', 'visible');
                }
                if (cursorFollowDist > distance) {
                    events_plot.addCursor(oldInfo.cursor);
                } else {
                    events_plot.hideCursor(chartObj.eventHandler.trackerElement);
                }
                return;
            }
        }

        var seriesLen = tooltipShared ? chartObj.seriesdata.length : 1;
        //highlighting new Info
        wrap.selectAll('#zcscatterhighlight')
            .selectAll('*')
            .remove();
        wrap.selectAll('#zcareahighlight')
            .selectAll('*')
            .remove();
        var tooltipSharedInfo = d3_map();

        for (var i = 0; i < seriesLen; i++) {
            var curSeriespos = tooltipShared ? i : seriespos;
            hoverSeries(
                chartObj,
                curSeriespos,
                xdata,
                itemPos,
                subseriesindex,
                nearestPoint,
                tooltipShared,
                cursorFollowDist > distance,
                tooltipSharedInfo
            );
        }

        //tooltip
        if (tooltipEnabled) {
            var tooltipPos = getTooltipPosition(
                    evt,
                    relmousePt,
                    xval,
                    yval,
                    chartObj,
                    nearestPoint,
                    axisrotated,
                    tooltipShared,
                    tooltipSharedInfo,
                    false,
                    isPolarAxisCategory,
                    isMultiPieChartType(chartObj.chartTypes[seriespos].value) &&
                        chartObj.subrenderer[subseriesindex][bubblepieUtils.getKey(data, seriespos, chartObj)].fillColorFunc
                ),
                tooltipArgs = tooltipPos.helpers;
            if (chartObj.dataObject.isBubbleEnabled) {
                tooltipPos.distance =
                    chartObj.dataObject.getXYvalue(axisrotated, stacked(type), xscale, ysc, yscindex, data, 'z', chartObj.zscale) / 2;
            }
            tooltip.show(chartObj, tooltipPos, null, true, tooltipArgs);

            oldInfo = chartObj.eventHandler.oldInfo;
            if (oldInfo) {
                oldInfo.containerLeft = cur_container_left;
                oldInfo.containerTop = cur_container_top;
            }
        }
    }

    function hoverPolygon(
        areahighlightWrap,
        type,
        plotOptions,
        commonPlotOptions,
        seriespos,
        subseriesindex,
        itemPos,
        xscale,
        ysc,
        yscindex,
        rangeBand,
        poly,
        tooltipShared
    ) {
        var seriesdata = chartObj.seriesdata;
        var fillOpacity = chartObj.systemConf.legend.patternPallete ? 0.5 : 0.2,
            fillColor = getRawColor(chartObj, seriespos, getCurrentDataPoint(seriespos, subseriesindex, itemPos), seriespos);
        var highlightEffect = chartObj.legendHighlight.getLegendHighlightEffect(seriespos, 'selectedSeries', 'plot');
        var highlightStyle =
            highlightEffect === NONE
                ? {
                      fill: TRANSPARENT
                  }
                : chartObj.legendHighlight.getEffect(
                      highlightEffect,
                      'area',
                      fillColor,
                      { fill: TRANSPARENT, stroke: fillColor },
                      seriespos,
                      'plot'
                  );
        if (!tooltipShared) {
            //For Shared tooltip, no need to remove adjacent highlighted wraps
            hideUnselectedWrap(wrap.selectAll('g#zcbarhighlight'));
        }
        areahighlightWrap
            .style('display', '')
            .attr('transform', `translate(${plotarea.left},${plotarea.top})`)
            .styles(highlightStyle);
        var interpolation = gvpFn(plotOptions, commonPlotOptions)('mode'),
            fillOpacity = validateNullVal(commonPlotOptions.hoveroptions.fillOpacity, 0.2);
        var y0 = null;
        var x1, y1, x2, y2;
        var x, y, width, height;
        var isCatNullX = isCatNull(dataset, 'x'),
            isCatNullY = isCatNull(dataset, 'y', yscindex);

        if (interpolation == 'linear' && poly != null && !isPolarAxisCategory) {
            var polyPoints = d3_merge(
                poly.map(function(d, i) {
                    return [d.x, d.y];
                })
            );

            var polygon = appendEle(areahighlightWrap, 'polygon', [poly], '.', 'class', `highlight${seriespos}`);
            polygon.attr('points', polyPoints.toString()).styles({
                'fill-opacity': fillOpacity,
                display: ''
            });
        } else {
            var isStkd = stacked(type),
                hasNegValue = chartObj.dataObject.containsNegativeData(yscindex),
                xaxisReversed = chartObj.systemConf.chart.axes.xaxis.reversed,
                yaxisReversed = chartObj.systemConf.chart.axes.yaxis[yscindex || 0].reversed,
                area,
                serdata;
            if (isPolarAxisCategory) {
                var renderer = chartObj.renderer[seriespos];
                serdata =
                    renderer.seriesdataAfterUpdate != null
                        ? renderer.seriesdataAfterUpdate[subseriesindex]
                        : seriesdata[seriespos].data[subseriesindex];
                var polarBaseIdx = yaxisReversed && !isStkd ? ysc.domain().length - 1 : 0;

                var outerRadiusFn = function(d, j) {
                        var OR = isStkd
                            ? ysc(d.y0 + d.y)
                            : ysc(dataset.getY(d, yscindex)) + (ysc.bandwidth != null ? ysc.bandwidth() / 2 : 0);
                        return mathMax(OR, ysc.range()[polarBaseIdx]);
                    },
                    innerRadiusFn = function(d, j) {
                        var IR;
                        if (isStkd) {
                            IR = seriespos == 0 ? mathMin(ysc.range()[0], ysc(hasNegValue)) : ysc(d.y0);
                        } else {
                            IR = ysc(ysc.domain()[polarBaseIdx]) + (ysc.bandwidth != null ? ysc.bandwidth() / 2 : 0);
                        }
                        return mathMax(IR, ysc.range()[polarBaseIdx]);
                    },
                    isLineHighlight = highlightEffect === 'strokeEnlarge';

                area = (isLineHighlight ? d3_lineRadial : d3_areaRadial)()
                    .angle(function(d, j) {
                        return xscale(dataset.getX(d));
                    })
                    .curve(d3_curves(interpolation, axisrotated));

                if (isLineHighlight) {
                    area.radius(outerRadiusFn);
                } else {
                    area.outerRadius(outerRadiusFn).innerRadius(innerRadiusFn);
                }
            } else {
                var isRangeChart = type == charttype.arearange;
                var yrangePos = isRangeChart ? 1 : null;
                var renderer = chartObj.renderer[seriespos];
                var currentSeriesdata =
                    renderer.seriesdataAfterUpdate != null
                        ? renderer.seriesdataAfterUpdate[subseriesindex]
                        : seriesdata[seriespos].data[subseriesindex];
                var data = currentSeriesdata[itemPos];
                var xval = chartObj.dataObject.getXYvalue(axisrotated, isStkd, xscale, ysc, yscindex, data, 'x', null, yrangePos),
                    yval = chartObj.dataObject.getXYvalue(axisrotated, isStkd, xscale, ysc, yscindex, data, 'y', null, yrangePos),
                    isNotFirstData = itemPos != 0,
                    isNotLastData = itemPos != currentSeriesdata.length - 1;

                if (xaxisReversed ? isNotLastData : isNotFirstData) {
                    var prevData = currentSeriesdata[xaxisReversed ? itemPos + 1 : itemPos - 1];
                    var prevXval = chartObj.dataObject.getXYvalue(
                        axisrotated,
                        isStkd,
                        xscale,
                        ysc,
                        yscindex,
                        prevData,
                        'x',
                        null,
                        yrangePos
                    );
                    var prevYval = chartObj.dataObject.getXYvalue(
                        axisrotated,
                        isStkd,
                        xscale,
                        ysc,
                        yscindex,
                        prevData,
                        'y',
                        null,
                        yrangePos
                    );
                    x1 = (xval + prevXval) / 2;
                    y1 = (yval + prevYval) / 2;
                }
                if (xaxisReversed ? isNotFirstData : isNotLastData) {
                    var nextData = currentSeriesdata[xaxisReversed ? itemPos - 1 : itemPos + 1];
                    var nextXval = chartObj.dataObject.getXYvalue(
                        axisrotated,
                        isStkd,
                        xscale,
                        ysc,
                        yscindex,
                        nextData,
                        'x',
                        null,
                        yrangePos
                    );
                    var nextYval = chartObj.dataObject.getXYvalue(
                        axisrotated,
                        isStkd,
                        xscale,
                        ysc,
                        yscindex,
                        nextData,
                        'y',
                        null,
                        yrangePos
                    );
                    x2 = (xval + nextXval) / 2;
                    y2 = (yval + nextYval) / 2;
                }

                x = axisrotated ? 0 : x1 == null ? mathMin(xval, x2) : x2 == null ? mathMin(x1, xval) : mathMin(x1, x2); //#ZC1632
                y = d3_min([y1, yval, y2, ysc(ysc.domain()[yaxisReversed ? 0 : 1])]);
                width = mathAbs(
                    axisrotated
                        ? d3_max([x1, xval, x2, ysc(hasNegValue) || xval])
                        : x1 != null && x2 != null
                        ? x2 - x1
                        : x1 == null
                        ? x2 == null
                            ? 1
                            : x2 - xval
                        : xval - x1
                );
                height = axisrotated
                    ? y1 == null
                        ? mathAbs(yval - y2)
                        : y2 == null
                        ? mathAbs(y1 - yval)
                        : mathAbs(y1 - y2)
                    : plotarea.height - y;

                var dx1 = Math.max(0, -x),
                    dy1 = Math.max(0, -y),
                    dx2 = Math.max(0, x + width - plotarea.width - (chartObj.scrolledX || 0)),
                    dy2 = Math.max(0, y + height - plotarea.height - (chartObj.scrolledY || 0));

                x = Math.max(0, x);
                y = Math.max(0, y);
                width -= dx1 + dx2;
                height -= dy1 + dy2;

                areahighlightWrap.style('display', 'block');

                serdata = currentSeriesdata;
                area = d3_area()
                    .defined(function(d, j) {
                        return !isCatNullY(dataset.getY(d, yscindex)) && !isCatNullX(dataset.getX(d));
                    })
                    // .x(function(d, j) {
                    //     return chartObj.dataObject.getXYvalue(axisrotated, isStkd, xscale, ysc, yscindex, d, "x")
                    // })
                    // .y(function(d, j) {
                    //     return chartObj.dataObject.getXYvalue(axisrotated, isStkd, xscale, ysc, yscindex, d, "y", null, yrangePos)
                    // })
                    .curve(d3_curves(interpolation, axisrotated));
                if (axisrotated) {
                    area.x0(function(d, i) {
                        if (isStkd) {
                            return seriespos == 0 && subseriesindex == 0 && i == 0 ? mathMax(ysc.range()[0], ysc(hasNegValue)) : ysc(d.y0);
                        } else {
                            return ysc.bandwidth != null ? ysc.origRange[0] + ysc.bandwidth() : mathMax(ysc.range()[0], ysc(hasNegValue));
                        }
                    })
                        .x1(function(d, j) {
                            return chartObj.dataObject.getXYvalue(axisrotated, isStkd, xscale, ysc, yscindex, d, 'x');
                        })
                        .y(function(d, j) {
                            return chartObj.dataObject.getXYvalue(axisrotated, isStkd, xscale, ysc, yscindex, d, 'y', null, yrangePos);
                        });
                } else {
                    area.x(function(d, j) {
                        return chartObj.dataObject.getXYvalue(axisrotated, isStkd, xscale, ysc, yscindex, d, 'x');
                    })
                        .y0(function(d, i) {
                            if (isStkd) {
                                return seriespos == 0 && subseriesindex == 0 && i == 0
                                    ? mathMin(ysc.range()[0], ysc(hasNegValue))
                                    : ysc(d.y0);
                            } else {
                                if (isRangeChart) {
                                    return chartObj.dataObject.getXYvalue(axisrotated, isStkd, xscale, ysc, yscindex, d, 'y', null, 0);
                                } else {
                                    return ysc.bandwidth != null
                                        ? ysc.origRange[0] + ysc.bandwidth()
                                        : mathMin(ysc.range()[0], ysc(hasNegValue));
                                }
                            }
                            // return (ysc.rangeBand != null) ? (ysc.range()[0]+ysc.rangeBand()) : Math.min(ysc.range()[0],ysc(hasNegValue));}
                        })
                        .y1(function(d, j) {
                            return chartObj.dataObject.getXYvalue(axisrotated, isStkd, xscale, ysc, yscindex, d, 'y', null, yrangePos);
                        });
                }
            }

            var properties = {
                    name: 'rect',
                    id: chartObj.id,
                    x: x,
                    y: y,
                    width: width,
                    height: height
                },
                args = {
                    chartObj: chartObj
                };
            var clipID = effects.get(effects.svg_clips, 'rect', properties, args);
            var areahighlight = appendEle(wrap.selectAll('#zcareahighlight'), 'path', [1], '.', 'class', `highlight${seriespos}`);
            areahighlight
                .datum(serdata)
                .attr('d', area)
                .attr('stroke', '')
                .attr('clip-path', isPolarAxisCategory ? null : `url(${chartObj.defsLocation}#${clipID})`)
                .styles({
                    'fill-opacity': fillOpacity,
                    display: ''
                });
        }
    }

    function hoverRect(barWrap, highlightpoint, radius, renderer, yaxisReversed, tooltipShared) {
        var t = getRectTranslation(highlightpoint),
            path = getRect(highlightpoint, radius, yaxisReversed);

        if (!tooltipShared) {
            //For Shared tooltip, no need to remove adjacent highlighted wraps
            hideUnselectedWrap(wrap.selectAll('g#zcareahighlight,g#zcscatterhighlight'));
        }
        wrap.selectAll('g#zcbarhighlight').style('display', '');
        barWrap
            .attr('d', path)
            .attr('transform', `translate(${t[0]},${t[1]})`)
            .style('display', '');
    }

    function hoverScatter(
        scatterWrap,
        type,
        plotOptions,
        commonPlotOptions,
        xval,
        yval,
        xscale,
        ysc,
        yscindex,
        seriesdata,
        subseriesindex,
        seriespos,
        itemPos,
        isAreaChart,
        ganttMarkerProp,
        tooltipShared
    ) {
        var isGantt = type == charttype.gantt,
            highlightEffect = chartObj.legendHighlight.getLegendHighlightEffect(seriespos, 'selectedSeries', 'plot'),
            isPatternPalleteEnabled =
                L_H.hasPatternPallete(chartObj) &&
                (isBubbleChartType(type) || type === charttype.scatter) &&
                highlightEffect !== 'strokeEnlarge',
            fillColor = getRawColor(
                chartObj,
                isMultiColoring(commonPlotOptions, chartObj) ? itemPos : seriespos,
                getCurrentDataPoint(seriespos, subseriesindex, itemPos),
                seriespos
            );
        if (!isGantt && !tooltipShared) {
            //For Shared tooltip, no need to remove adjacent highlighted wraps
            hideUnselectedWrap(wrap.selectAll('g#zcbarhighlight'));
        }
        if (!isAreaChart && !isPolarAxisCategory) {
            hideUnselectedWrap(wrap.selectAll('g#zcareahighlight').selectAll('*'));
            hideUnselectedWrap(wrap.selectAll('g#zcareahighlight'));
        }

        var scatterParentWrap = wrap.selectAll('#zcscatterhighlight'),
            mOptions = plotOptions
                ? mergeJSON(commonPlotOptions.hoveroptions, plotOptions.hoveroptions, true)
                : commonPlotOptions.hoveroptions,
            innerSymbolSize,
            outerSymbolSize;

        // fix the marker
        if (!defined(mOptions.symbol)) {
            mOptions.symbol = pick(getPropVal(plotOptions, 'marker.symbol'), getPropVal(commonPlotOptions, 'marker.symbol'));
        }

        if (isGantt && commonPlotOptions.marker && commonPlotOptions.marker.enabled === false) {
            return;
        }

        if (isBubbleChartType(type)) {
            var renderer = chartObj.renderer[seriespos],
                currentSeriesdata = defined(renderer.seriesdataAfterUpdate)
                    ? renderer.seriesdataAfterUpdate[subseriesindex]
                    : seriesdata[seriespos].data[subseriesindex],
                z = chartObj.dataObject.getXYvalue(
                    axisrotated,
                    stacked(type),
                    xscale,
                    ysc,
                    yscindex,
                    currentSeriesdata[itemPos],
                    'z',
                    chartObj.zscale
                );

            if (isPatternPalleteEnabled) {
                outerSymbolSize = z * 0.1;
                innerSymbolSize = 0;
            } else {
                outerSymbolSize = (z + perToPx('2.5%', mathMin(plotarea.width, plotarea.height))) * 0.1;
                innerSymbolSize = (z / 2) * 0.1;
            }
        } else {
            var lineOrArea = type === charttype.line || isAreaChartType(type),
                markerSize =
                    plotOptions && plotOptions.marker
                        ? gvpFn(plotOptions.marker, commonPlotOptions.marker)('size')
                        : commonPlotOptions.marker.size;

            var symbolsize =
                (lineOrArea || isPolarAxisCategory) && (markerSize == null || markerSize === '')
                    ? gvpFn(plotOptions, commonPlotOptions)('strokeWidth')
                    : markerSize || 1;
            symbolsize = mOptions.size != null && mOptions.size !== '' ? 1 : +symbolsize;
            if (isGantt) {
                mOptions = plotOptions
                    ? mergeJSON(ganttMarkerProp.hoveroptions, plotOptions.hoveroptions, true)
                    : ganttMarkerProp.hoveroptions;
            }

            var outerSymbolSizeFactor, innerSymbolSizeFactor;

            if (isPolarAxisCategory) {
                outerSymbolSizeFactor = 0.7;
                innerSymbolSizeFactor = 0.35;
            } else if (lineOrArea) {
                outerSymbolSizeFactor = 0.9;
                innerSymbolSizeFactor = 0.4;
            } else if (isPatternPalleteEnabled) {
                outerSymbolSizeFactor = 1;
                innerSymbolSizeFactor = 0;
            } else if (LOCATION_MARKERS.has(mOptions.symbol)) {
                outerSymbolSizeFactor = 1;
                innerSymbolSizeFactor = 0.5;
            } else {
                outerSymbolSizeFactor = 2;
                innerSymbolSizeFactor = 1;
            }

            outerSymbolSize = symbolsize * outerSymbolSizeFactor;
            innerSymbolSize = symbolsize * innerSymbolSizeFactor;
        }

        scatterParentWrap.attr('transform', `translate(${plotarea.left},${plotarea.top})`).style('display', '');
        addSVGmarker(
            scatterWrap,
            null,
            isPatternPalleteEnabled ? { symbol: mOptions.symbol } : mOptions,
            {
                color: isPatternPalleteEnabled ? null : fillColor,
                symbolSize: {
                    inner: innerSymbolSize,
                    outer: outerSymbolSize
                },
                transform: [xval, yval],
                imageSize: 15
            },
            chartObj
        );

        if (isPatternPalleteEnabled) {
            var highlightStyle =
                highlightEffect === NONE
                    ? {
                          fill: TRANSPARENT
                      }
                    : chartObj.legendHighlight.getEffect(
                          highlightEffect,
                          'scatter',
                          HIGHLIGHT_EFFECT_REGEX.test(highlightEffect) ? fillColor : TRANSPARENT,
                          { fill: TRANSPARENT, stroke: fillColor },
                          seriespos,
                          'plot'
                      );
            scatterWrap.styles(highlightStyle);
        }
    }
    // will use it on mapEvents
    events_plot.hoverScatter = hoverScatter;

    function hoverSeries(
        chartObj,
        seriespos,
        actXdata,
        itemPos,
        subseriesindex,
        nearestPoint,
        tooltipShared,
        showCursor,
        tooltipSharedInfo,
        highlightPoints
    ) {
        var seriesdata = chartObj.seriesdata;
        var yscindex = getParsedYAxisOrder(seriesdata[seriespos].yaxiscolumnorder);
        var ysc = yscale[yscindex].scale;
        var type = chartObj.chartTypes[seriespos].value;
        var isRangeChart = type == charttype.arearange;
        var isGantt = type === charttype.gantt;
        var yrangePos = isRangeChart || isGantt ? 1 : null;
        var rangeBand = xscale.bandwidth != null ? xscale.bandwidth() : null;
        var renderer = chartObj.renderer[seriespos];
        var currentSeriesdata =
            renderer.seriesdataAfterUpdate != null
                ? renderer.seriesdataAfterUpdate[subseriesindex]
                : seriesdata[seriespos].data[subseriesindex];
        if (currentSeriesdata == null) {
            return;
        }
        var actData = currentSeriesdata[itemPos];
        itemPos = actData && dataset.getActualX(actData) === actXdata ? itemPos : getContinousDataItempos(actXdata, currentSeriesdata);
        if (itemPos == null || eventDisabled(chartObj.seriesdata[seriespos]) || isNullData(currentSeriesdata[itemPos], dataset, yscindex)) {
            // If there is no data for this particular item for this series
            return;
        }

        tooltipSharedInfo.set(seriespos, itemPos);
        tooltip_content.updateSharedInfo(chartObj, tooltipSharedInfo, seriespos);

        var data = currentSeriesdata[itemPos];
        var isAreaChart;

        var xval = chartObj.dataObject.getXYvalue(axisrotated, stacked(type), xscale, ysc, yscindex, data, 'x', null, yrangePos);
        var yval = chartObj.dataObject.getXYvalue(axisrotated, stacked(type), xscale, ysc, yscindex, data, 'y', null, yrangePos);

        var plotOptions = seriesdata[seriespos].plotoptions;
        var commonPlotOptions = systemConf.chart.plot.plotoptions[charttypenames.get(type)],
            eventType = chartObj.zcEventType || 'mousemove';
        var mouseMoveFn = getEventConf(plotOptions, commonPlotOptions, chartObj, eventType),
            cursor = getEventConf(plotOptions, commonPlotOptions, chartObj, 'cursor');

        if (showCursor) {
            events_plot.addCursor(cursor);
        } else {
            events_plot.hideCursor(chartObj.eventHandler.trackerElement);
        }

        // handle user action
        var event = d3_event() || chartObj.eventHandler.targetEvent;
        if (defined(mouseMoveFn)) {
            triggerUserCallback(mouseMoveFn, nearestPoint, event, { x: xval, y: yval });
        }

        /**
         *
         * if event.preventDefault() is called, then
         * it does not do default action.
         */
        if (isDefaultAllowed(mouseMoveFn, event)) {
            highlightpoint = highlightPoints || highlightpoint;
            if ((highlightpoint != null || tooltipShared) && isBarChartType(type)) {
                var yaxisReversed = chartObj.systemConf.chart.axes.yaxis[yscindex || 0].reversed;
                var borderOptions = gvpFn(plotOptions, commonPlotOptions)('border') || {};
                var borderRadius = type === charttype.bullet ? 0 : borderOptions.radius;
                borderRadius = borderOptions && borderOptions.show ? borderRadius : 0;
                var renderer = chartObj.renderer[seriespos];
                //radius = renderer.getParamValue("radius",null,null,null,borderRadius);
                var radius = chartObj.dataObject.getTransRectValue('radius', null, null, null, null, borderRadius);
                var highlightEffect = chartObj.legendHighlight.getLegendHighlightEffect(seriespos, 'selectedSeries', 'plot'),
                    unHighlightEffect = chartObj.legendHighlight.getLegendHighlightEffect(seriespos, 'unselectedSeries', 'plot'),
                    fillColor = getRawColor(
                        chartObj,
                        isMultiColoring(commonPlotOptions, chartObj) ? itemPos : seriespos,
                        getCurrentDataPoint(seriespos, subseriesindex, itemPos),
                        seriespos
                    ),
                    currentColor =
                        HIGHLIGHT_EFFECT_REGEX.test(highlightEffect) || UNHIGHLIGHT_EFFECT_REGEX.test(unHighlightEffect)
                            ? fillColor
                            : TRANSPARENT;

                var highlightStyle =
                    highlightEffect === NONE
                        ? {
                              fill: currentColor
                          }
                        : chartObj.legendHighlight.getEffect(
                              highlightEffect,
                              'bar',
                              currentColor,
                              { fill: currentColor, stroke: fillColor },
                              seriespos,
                              'plot'
                          );

                var barWrap = appendEle(
                    wrap.selectAll('#zcbarhighlight'),
                    'path',
                    [1],
                    '.',
                    'class',
                    `highlight${seriespos}${subseriesindex}`
                )
                    // var barWrap = wrap.selectAll("#zcbarhighlight").select("#barpath")
                    .styles(highlightStyle);

                // if(tooltipShared && !isHeatMapChartType(chartObj.chartTypes[seriespos].value)) { //For shared tooltip get highlighlight position on demand for every series
                //     var seriesdataAfterUpdate = renderer.seriesdataAfterUpdate[subseriesindex][itemPos];
                //     var commonOption = renderer.commonOption || getBarCommonOption(chartObj.rendererConf[seriespos], seriespos);
                //     highlightpoint = getBarPosition(chartObj, seriesdataAfterUpdate, subseriesindex, commonOption);
                // }

                if (tooltipShared && !isHeatMapChartType(chartObj.chartTypes[seriespos].value)) {
                    //For shared tooltip get highlighlight position on demand for every series
                    //For shared tooltip get highlight position on demand for every series
                    highlightpoint = getPointRectangle(seriespos, subseriesindex, itemPos);
                }

                hoverRect(barWrap, highlightpoint, radius, renderer, yaxisReversed, tooltipShared);
                var levelMarker =
                    plotOptions && plotOptions.levelMarker
                        ? mergeJSON(commonPlotOptions.levelMarker, plotOptions.levelMarker, true)
                        : commonPlotOptions.levelMarker;
                if (isGantt && levelMarker.enabled) {
                    var levelMarkrad = chartObj.dataObject.getTransRectValue('radius', null, null, null, 0);
                    var style = {
                        fill: levelMarker.color || fillColor
                    };
                    var levelMarkerWrap = appendEle(wrap.selectAll('#zcbarhighlight'), 'path', [1], '#', 'id', 'levelMarkerpath');
                    levelMarkerWrap.styles(style);
                    hoverRect(levelMarkerWrap, levelMarkerPoint, levelMarkrad, renderer, yaxisReversed);
                }
            }

            if (isAreaChartType(type) || isPolarAxisCategory) {
                isAreaChart = true;
                var areahighlightWrap = wrap.selectAll('#zcareahighlight');
                hoverPolygon(
                    areahighlightWrap,
                    type,
                    plotOptions,
                    commonPlotOptions,
                    seriespos,
                    subseriesindex,
                    itemPos,
                    xscale,
                    ysc,
                    yscindex,
                    rangeBand,
                    nearestPoint.poly,
                    tooltipShared
                );
            }
            if (isAreaChartType(type) || isPointChartType(type) || type === charttype.line || isPolarAxisCategory || isGantt) {
                var scatterWrap = wrap.selectAll('#zcscatterhighlight');
                var currentWrap = appendEle(scatterWrap, 'g', [1], '.', 'class', `highlight${seriespos}`);
                var ganttMarkerSize = 0;

                if (isRangeChart || isGantt) {
                    var yrangePos_0 = appendEle(currentWrap, 'g', [1], '.', 'class', 'yrangePos_0');
                    var yrangePos_1 = appendEle(currentWrap, 'g', [1], '.', 'class', 'yrangePos_1');
                    var xval_0,
                        yval_0,
                        prop = {};
                    if (isGantt) {
                        prop = renderer.markerProperties;

                        var pos = renderer.getMarkerPosition(data, prop.subseriesIndex, 0, renderer);
                        if (axisrotated) {
                            xval = pos.y;
                            yval = pos.x + pos.width / 2;
                            xval_0 = pos.y + pos.height;
                            yval_0 = pos.x + pos.width / 2;
                        } else {
                            xval = pos.x + pos.width / 2;
                            yval = pos.y;
                            xval_0 = pos.x + pos.width / 2;
                            yval_0 = pos.y + pos.height;
                        }
                    } else {
                        xval_0 = chartObj.dataObject.getXYvalue(axisrotated, stacked(type), xscale, ysc, yscindex, data, 'x', null, 0);
                        yval_0 = chartObj.dataObject.getXYvalue(axisrotated, stacked(type), xscale, ysc, yscindex, data, 'y', null, 0);
                    }
                    hoverScatter(
                        yrangePos_0,
                        type,
                        plotOptions,
                        commonPlotOptions,
                        xval_0,
                        yval_0,
                        xscale,
                        ysc,
                        yscindex,
                        seriesdata,
                        subseriesindex,
                        seriespos,
                        itemPos,
                        isAreaChart,
                        {
                            size: prop.size,
                            hoveroptions: commonPlotOptions.hoveroptions.end
                        }
                    );
                    hoverScatter(
                        yrangePos_1,
                        type,
                        plotOptions,
                        commonPlotOptions,
                        xval,
                        yval,
                        xscale,
                        ysc,
                        yscindex,
                        seriesdata,
                        subseriesindex,
                        seriespos,
                        itemPos,
                        isAreaChart,
                        {
                            size: prop.size,
                            hoveroptions: commonPlotOptions.hoveroptions.start
                        }
                    );
                } else {
                    hoverScatter(
                        currentWrap,
                        type,
                        plotOptions,
                        commonPlotOptions,
                        xval,
                        yval,
                        xscale,
                        ysc,
                        yscindex,
                        seriesdata,
                        subseriesindex,
                        seriespos,
                        itemPos,
                        isAreaChart,
                        null,
                        tooltipShared
                    );
                }
            }

            if (isPolarAxisCategory) {
                wrap.selectAll('#zcareahighlight,#zcscatterhighlight')
                    .attrs({
                        'clip-path': null,
                        transform: `translate(${plotarea.left + chartObj.plotarea_approx.x},${plotarea.top + chartObj.plotarea_approx.y})`
                    })
                    .style('display', '');
            }
        }
        nearestPoint.data = data;
        chartObj.eventHandler.oldInfo = {
            nearestPoint: nearestPoint,
            type: type,
            chartObj: chartObj,
            isNoInfo: false,
            cursor: cursor
        };
    }

    function getOrdinalDataPoint(relmousePt, xscale, yscale, systemConf, isMoving) {
        var xaxisReversed = chartObj.systemConf.chart.axes.xaxis.reversed,
            onlyBarCombo_noAreaChart =
                chartObj.dataObject.noofBarEnabledSeries && !chartObj.dataObject.noofAreaEnabledSeries && tooltipShared;
        //only for bar charts, bar charts with line or scatter, need to find invert x Pos with useWithinBound or withinRangebandoption.
        var categorypos = scale.invert(ORDINAL_DATATYPE, relmousePt, barscale1, axisrotated, null, xaxisReversed, onlyBarCombo_noAreaChart);
        var seriesdata = chartObj.seriesdata;

        var highlightpoint,
            levelMarkerPoint,
            itemPos = {},
            minTriggerOffset,
            xpos,
            subseries = {},
            seriespos = {},
            seriespos_cached,
            xpos_shared = [],
            triggerOffset_ChartType = {};
        var poly = null;
        var containsLine = false;
        var toHighlight = false,
            sampleTaken = {},
            currentQualitativeRangePoint = null,
            is_within_bubble,
            prevBubbleInfo;

        var X = 'x',
            Y = 'y';
        if (axisrotated) {
            X = 'y';
            Y = 'x';
        }

        var yminimumBisected = seriesdata.map(function(d, i) {
            var type = chartObj.chartTypes[i].value;
            minTriggerOffset =
                minTriggerOffset == null || chartObj.dataObject.isCombinationalChart
                    ? triggerOffset_ChartType[type] || getTriggeringOffset(chartObj, type, null)
                    : minTriggerOffset; //#ZC1349

            var yaxiscolumnorder = getParsedYAxisOrder(d.yaxiscolumnorder);
            var yaxisReversed = chartObj.systemConf.chart.axes.yaxis[yaxiscolumnorder || 0].reversed;
            var ysc = yscale[yaxiscolumnorder].eventscale;
            var hasNegValue = chartObj.dataObject.containsNegativeData(yaxiscolumnorder),
                isCatNullY = isCatNull(dataset, 'y', yaxiscolumnorder),
                polygonPoints = {},
                plotOptions = d.plotoptions,
                commonPlotOptions = systemConf.chart.plot.plotoptions[chartObj.chartTypes[i].name],
                markerOptions =
                    (plotOptions ? mergeJSON(commonPlotOptions.marker, plotOptions.marker, true) : commonPlotOptions.marker) || {},
                defaultMarkerSize = (markerOptions.enabled ? markerOptions.size || 1 : 1) * 10;

            if (eventDisabled(d)) {
                return [null];
            } else {
                var renderer = chartObj.renderer[i];

                /**
                 *
                 * @type {*|{stacked, categories, axisrotated, bandwidth1, bandwidth2, barDatasetIndex, padding, yscale, hasNegValue, yaxiscolumnorder, addRangeBand, barWidth, levelMarkerWidth, levelMarkerPadding}|*}
                 */
                var commonOption = renderer.commonOption || getBarCommonOption(chartObj.rendererConf[i], i);

                if (!events_plot.bulletEvents && charttype.bullet === type) {
                    events_plot.bulletEvents = new events.BulletEvent(chartObj, renderer.commonOption, xscale, yscale[0].scale);
                }

                var bulletEvent = events_plot.bulletEvents;

                return d.data.map(function(d, j) {
                    // var curpt = {};
                    var seriesdataAfterUpdate = renderer.seriesdataAfterUpdate != null ? renderer.seriesdataAfterUpdate[j] : d;
                    xpos = getCategoryXpos(categorypos, seriesdataAfterUpdate) || null; //fix for different category set for two different series
                    if (xpos == null) {
                        return;
                    }

                    if (tooltipShared) {
                        //For shared tooltip, no need to calculate distance based on data. Should highlight all data for xpos
                        var isWithinClrRange = isWithinColorRange(dataset.getClr(seriesdataAfterUpdate[xpos[0]]), chartObj);
                        // itemPos[type] = categorypos;
                        xpos_shared[i] = xpos_shared[i] || [];
                        xpos_shared[i][j] = xpos;
                        dist = isWithinClrRange ? 0 : null;
                        toHighlight = true;
                        if (isBarChartType(type)) {
                            seriespos_cached =
                                seriespos_cached ||
                                scale.ordinalSubInvert(relmousePt, itemPos[type], barscale2, barscale1, axisrotated, type, xaxisReversed);
                            seriespos.shared = seriespos_cached;
                            itemPos.shared = categorypos;
                            seriespos.shared = i; //chartObj.userdata.seriesdata.stackGroup ? chartObj.userdata.seriesdata.stackGroup[seriespos][0] : seriespos; //stackgrouping with sharedtooltip ZC1366
                        } else {
                            seriespos[type] = i;
                            itemPos[type] = categorypos;
                        }
                        return dist;
                    }
                    if (isHeatMapChartType(type)) {
                        var a = null;
                        seriesdataAfterUpdate.forEach(function(dt, k) {
                            var curpt = getHeatCellPosition(chartObj, dt, xscale, ysc, yaxiscolumnorder);
                            var isWithinClrRange = isWithinColorRange(dataset.getClr(dt), chartObj);
                            var dist = isWithin(curpt, relmousePt);

                            if (!isNullData(dt, dataset, yaxiscolumnorder) && isWithinClrRange && dist == 0 && a === null) {
                                toHighlight = true;
                                itemPos[type] = k;
                                seriespos[type] = i;
                                highlightpoint = curpt;
                                a = dist;
                            }
                        });
                        return a;
                    } else if (isBarChartType(type)) {
                        var len = xpos.length;
                        var dist = null;
                        for (var m = 0; m < len; m++) {
                            var curXpos = xpos[m]; //Handled multi data for same category
                            if (seriesdataAfterUpdate[curXpos] == null) {
                                return null;
                            }

                            var curpt = getBarPosition(chartObj, seriesdataAfterUpdate[curXpos], j, commonOption);

                            if (
                                isCatNullY(seriesdataAfterUpdate[curXpos]) ||
                                isNullData(seriesdataAfterUpdate[curXpos], dataset, yaxiscolumnorder)
                            ) {
                                //If the y value is null and nullformatter is undefined, don't show tooltip for that data
                                return null;
                            }

                            var curdist = isWithin(curpt, relmousePt, yaxisReversed);
                            var isWithinClrRange = isWithinColorRange(dataset.getClr(seriesdataAfterUpdate[curXpos]), chartObj);

                            if (isWithinClrRange && curdist == 0 && !sampleTaken[type]) {
                                dist = 0;
                                toHighlight = true;
                                sampleTaken[type] = true;
                                seriespos[type] = i;
                                itemPos[type] = curXpos;
                                highlightpoint = curpt;
                                if (type === charttype.gantt) {
                                    levelMarkerPoint = renderer.getBarLevelMarkerProp(
                                        seriesdataAfterUpdate[curXpos],
                                        j,
                                        curXpos,
                                        renderer,
                                        commonOption
                                    );
                                }
                            }

                            if (charttype.bullet === type) {
                                currentQualitativeRangePoint = bulletEvent.getCurrentPoint(relmousePt, seriesdataAfterUpdate[curXpos]);
                            }
                        }
                        return dist;
                    } else if (isAreaChartType(type)) {
                        xpos = xpos[0]; //Hack for other charts where multi data for same category not handled
                        var curpt = seriesdataAfterUpdate[xpos],
                            isWithinClrRange = isWithinColorRange(dataset.getClr(curpt), chartObj),
                            dist = null;

                        if (isNullData(curpt, dataset, yaxiscolumnorder)) {
                            return;
                        }

                        var isRangeChart = type == charttype.arearange;
                        polygonPoints = getPolygonPoints(
                            i,
                            j,
                            xpos,
                            seriesdataAfterUpdate,
                            seriesdataAfterUpdate[xpos - 1],
                            seriesdataAfterUpdate[xpos],
                            seriesdataAfterUpdate[xpos + 1],
                            chartObj.dataObject,
                            dataset,
                            axisrotated,
                            stacked(type),
                            xscale,
                            ysc,
                            yaxiscolumnorder,
                            hasNegValue,
                            isRangeChart
                        );
                        var isWithinPoly = pointInPolygon(relmousePt, polygonPoints.points);
                        if (isWithinPoly) {
                            dist =
                                j == 0
                                    ? distToSegment(relmousePt, polygonPoints.curpt, polygonPoints.curNext)
                                    : j == seriesdataAfterUpdate.length - 1
                                    ? distToSegment(relmousePt, polygonPoints.prevCur, polygonPoints.curpt)
                                    : d3_min([
                                          distToSegment(relmousePt, polygonPoints.prevCur, polygonPoints.curpt),
                                          distToSegment(relmousePt, polygonPoints.curpt, polygonPoints.curNext)
                                      ]);
                        }

                        if (isWithinClrRange && (minTriggerOffset == null || dist < minTriggerOffset) && dist != null) {
                            toHighlight = true;
                            sampleTaken[type] = true;
                            minTriggerOffset = dist;
                            seriespos[type] = i;
                            subseries[type] = j;
                            itemPos[type] = xpos;
                            poly = polygonPoints.points;
                            triggerOffset_ChartType[type] = minTriggerOffset;
                            return dist; // ZC1349
                        }
                        return null;
                    } else if (isPointChartType(type) || type === charttype.line) {
                        var len = xpos.length;
                        var dist = NULL;
                        for (var m = 0; m < len; m++) {
                            var curXpos = xpos[m], //Handled multi data for same category
                                curpt = seriesdataAfterUpdate[curXpos],
                                isWithinClrRange = isWithinColorRange(dataset.getClr(curpt), chartObj);

                            if (isNullData(curpt, dataset, yaxiscolumnorder) || !isWithinClrRange) {
                                return;
                            }

                            var cx = chartObj.dataObject.getXYvalue(axisrotated, stacked(type), xscale, ysc, yaxiscolumnorder, curpt, 'x'),
                                cy = chartObj.dataObject.getXYvalue(axisrotated, stacked(type), xscale, ysc, yaxiscolumnorder, curpt, 'y'),
                                shapeInfo = { cx: cx, cy: cy },
                                symbolbbox = renderer.getSymbolBBox && renderer.getSymbolBBox(curpt),
                                is_within_current_bubble,
                                condition = false,
                                curDist = NULL;

                            if (symbolbbox != NULL) {
                                var symbolSegment = constructSegmentAroundDataPoint(
                                    curpt,
                                    dataset,
                                    xscale,
                                    ysc,
                                    yaxiscolumnorder,
                                    type,
                                    NULL,
                                    X,
                                    Y,
                                    symbolbbox
                                );
                                curDist = distToSegment(relmousePt, symbolSegment[0], symbolSegment[1]);

                                var width = symbolbbox.width,
                                    height = symbolbbox.height,
                                    x = cx - width / 2,
                                    y = cy - height / 2;

                                shapeInfo.area = geometryUtils.areaOfRect(width, height);
                                is_within_current_bubble = geometryUtils.pointInRectangle(relmousePt, {
                                    x: x,
                                    y: y,
                                    width: width,
                                    height: height
                                });
                            } else {
                                curDist = getMinDistByPtToPt(curpt, relmousePt, xscale, ysc, yaxiscolumnorder, type, NULL, X, Y);
                                var r = isBubbleChartType(type)
                                    ? chartObj.dataObject.getXYvalue(
                                          axisrotated,
                                          stacked(type),
                                          xscale,
                                          ysc,
                                          yaxiscolumnorder,
                                          curpt,
                                          'z',
                                          chartObj.zscale
                                      )
                                    : defaultMarkerSize;
                                r /= 2;

                                shapeInfo.radius = r; // TODO: remove
                                shapeInfo.area = geometryUtils.areaOfCircle(r);
                                is_within_current_bubble = curDist <= r;
                            }

                            shapeInfo.dist = curDist;
                            is_within_bubble = is_within_bubble || is_within_current_bubble;

                            if (is_within_bubble) {
                                if (is_within_current_bubble) {
                                    if (prevBubbleInfo) {
                                        var prevSurface = prevBubbleInfo.area,
                                            curSurface = shapeInfo.area,
                                            distanceBetweenCircles = mathSQRT(sqr(prevBubbleInfo.cx - cx) + sqr(prevBubbleInfo.cy - cy)),
                                            inside =
                                                isBubbleChartType(type) &&
                                                distanceBetweenCircles * 0.9 <= mathAbs(prevBubbleInfo.radius - shapeInfo.radius);

                                        condition =
                                            prevBubbleInfo.dist == NULL ||
                                            curDist < prevBubbleInfo.dist ||
                                            !(prevSurface < curSurface && inside);
                                    } else {
                                        condition = true;
                                    }
                                }
                            } else {
                                condition = curDist <= minTriggerOffset;
                            }

                            if (type === charttype.bubblepie) {
                                var pt = isNullData(curpt, dataset, yaxiscolumnorder)
                                    ? {
                                          x: 0,
                                          y: 0
                                      }
                                    : {
                                          x: cx,
                                          y: cy
                                      };

                                condition = condition && pointInPie(relmousePt, pt, i, j, curpt);
                                if (condition && sampleTaken[type] && chartObj.chartTypes[seriespos.point].value === charttype.bubblepie) {
                                    //#ZC1323
                                    var key = bubblepieUtils.getKey(curpt, i, chartObj),
                                        prevKey = bubblepieUtils.getKey(
                                            getCurrentDataPoint(seriespos.point, subseries.point, itemPos.point),
                                            seriespos.point,
                                            chartObj
                                        );

                                    condition = condition && bubblePieKeys.indexOf(key) >= bubblePieKeys.indexOf(prevKey);
                                }
                            }

                            if (condition) {
                                toHighlight = true;
                                sampleTaken[type] = true;
                                dist = curDist;
                                minTriggerOffset = curDist;
                                seriespos.point = i;
                                subseries.point = j;
                                itemPos.point = curXpos;
                                containsLine = type === charttype.line || isPointChartType(type);
                                if (is_within_current_bubble) {
                                    prevBubbleInfo = shapeInfo;
                                }
                            }
                        }
                        return dist;
                    }
                });
            }
        });

        var subseriesindex = [],
            subseriesLength = [];
        var min_yminimumBisected = yminimumBisected.map(function(d, i) {
            var chartType = chartObj.chartTypes[i].value,
                type = isPointChartType(chartType) || chartType === $ZC.charttype.line ? 'point' : chartType,
                min = is_within_bubble && isBubbleChartType(chartType) && seriespos[type] === i ? d[subseries[type]] : d3_min(d); //Bubblepie with subseries

            if (min != null) {
                subseriesindex[i] = d.indexOf(min);
                subseriesLength[i] = d.length;
            } else {
                subseriesindex[i] = subseriesLength[i] = null;
            }
            return min;
        });
        if (containsLine) {
            min_yminimumBisected = min_yminimumBisected.map(function(d, i) {
                var type = chartObj.chartTypes[i].value;
                type = isPointChartType(type) || type === charttype.line ? 'point' : type;
                if (type === 'point' && seriespos[type] === i) {
                    return d;
                } else {
                    return null;
                }
            });
        }

        var actualSeriesPosIndex =
            tooltipShared && defined(seriespos_cached)
                ? seriespos.shared
                : isHeatMapChartType(chartObj.chartTypes[0].value)
                ? min_yminimumBisected.lastIndexOf(d3_min(min_yminimumBisected))
                : min_yminimumBisected.indexOf(d3_min(min_yminimumBisected)); //Why lastIndexOf - ZC1487
        if (min_yminimumBisected[actualSeriesPosIndex] == null || !toHighlight) {
            actualSeriesPosIndex = null;
        }

        if (defined(currentQualitativeRangePoint) && currentQualitativeRangePoint.type === 'marker') {
            // bullet target marker has more priority
            actualSeriesPosIndex = null;
        } else if (defined(actualSeriesPosIndex)) {
            currentQualitativeRangePoint = null;
        }
        var selectedType =
            tooltipShared && defined(seriespos_cached)
                ? 'shared'
                : defined(actualSeriesPosIndex)
                ? chartObj.chartTypes[actualSeriesPosIndex].value
                : null;
        selectedType = !tooltipShared && (isPointChartType(selectedType) || selectedType === charttype.line) ? 'point' : selectedType;
        return {
            seriesPos: selectedType && seriespos[selectedType],
            itemPos: selectedType && itemPos[selectedType],
            itemPosShared: xpos_shared,
            subseriesindex: subseriesindex,
            subseriesLength: subseriesLength,
            highlightpoint: highlightpoint || null,
            levelMarkerPoint: levelMarkerPoint || null,
            distance: min_yminimumBisected[seriespos],
            poly: poly,
            currentQualitativeRangePoint: currentQualitativeRangePoint,
            customPoint: CustomPoints_th ? CustomPoints_th.getClosest(relmousePt, minTriggerOffset) : null
        };
    }

    function getCategoryXpos(xpos, data, isMultiCat) {
        var categories = xscale.domain();
        var xval = categories[xpos],
            pos = null,
            isCatNullX = isCatNull(dataset, 'x');
        data.forEach(function(d, i) {
            var hasX = !isCatNullX(dataset.getX(d)) && dataset.getX(d) === xval;
            if (hasX) {
                pos == null ? (pos = [i]) : pos.push(i); // Create array for multiple data for same category
            }
            return hasX;
        });
        return pos;
    }

    function getContinousDataItempos(item, data) {
        var pos = null,
            isCatNullX = isCatNull(dataset, 'x');
        data.some(function(d, i) {
            var hasX = !isCatNullX(dataset.getX(d)) && dataset.getX(d) === item;
            if (hasX) {
                pos = i;
            }
            return hasX;
        });
        return pos;
    }

    function addNoteFunction(args, seriespos, itemPos, currentSeriesdata, xval, yval) {
        var canAddOnDoubleclickEvent = chartObj.systemConf.notes && chartObj.systemConf.notes.addOnDoubleclickEvent;
        if (annotation && canAddOnDoubleclickEvent) {
            if (!args) {
                var seriesdata = chartObj.seriesdata;
                var currentEle = currentSeriesdata[itemPos];
                var clonedata = currentEle.slice();
                var data = clonedata.splice(0, clonedata.length);
                if (seriesdata.length > 1) {
                    data.push(seriesdata[seriespos].seriesname);
                }
                args = {
                    coordinate: [xval, yval],
                    data: data,
                    basedOn: 'plot'
                };
            }
            annotation.addNotes(chartObj, args);
        }
    }

    function bisectX(relmousePt, xscale, yscale, systemConf) {
        var plotX = xscale.invert(axisrotated ? relmousePt.y : relmousePt.x);
        var seriesdata = systemConf.seriesdata.chartdata;

        var xBisector = d3_bisector(function(d) {
            return dataset.getX(d);
        }).left;

        var xbisect = seriesdata.map(function(d, i) {
            if (d.disabled) {
                return null;
            }

            var renderer = chartObj.renderer[i];
            return d.data.map(function(d, j) {
                d = renderer.seriesdataAfterUpdate != null ? renderer.seriesdataAfterUpdate[j] : d;
                return xBisector(d, plotX, 0, d.length - 1);
            });
        });

        var xminimumBisected = seriesdata.map(function(d, i) {
            if (d.disabled) {
                return null;
            }

            var renderer = chartObj.renderer[i];
            var yscindex = getParsedYAxisOrder(d.yaxiscolumnorder);
            var type = chartObj.chartTypes[i].value;
            var ysc = yscale[yscindex].scale;
            return d.data.map(function(d, j) {
                d = renderer.seriesdataAfterUpdate != null ? renderer.seriesdataAfterUpdate[j] : d;
                var xbis = xbisect[i][j],
                    curpt,
                    isWithinClrRange;

                if (d[xbis]) {
                    //Test_615.json If seriesdata.chartdata[i].data =[[]] and other series has data, d[xbis] will be null
                    curpt = {
                        x: xscale(dataset.getX(d[xbis])),
                        y: stacked(type) ? ysc(d[xbis].y + d[xbis].y0) : ysc(dataset.getY(d[xbis], yscindex))
                    };
                    isWithinClrRange = isWithinColorRange(dataset.getClr(d), chartObj);
                }

                if (xbis > 0) {
                    var dis = mathAbs(relmousePt.x - curpt.x);
                    var prevpt = {
                        x: xscale(dataset.getX(d[xbis - 1])),
                        y: stacked(type) ? ysc(d[xbis - 1].y + d[xbis - 1].y0) : ysc(dataset.getY(d[xbis - 1], yscindex))
                    };
                    var prevptDis = mathAbs(relmousePt.x - prevpt.x);
                    var vv = dis > prevptDis ? xbis - 1 : xbis;
                    dis = dis > prevptDis ? prevptDis : dis;
                    var curpt2mousedis = lineDistance(curpt, relmousePt);
                    var prept2mousedis = lineDistance(relmousePt, prevpt);
                    return {
                        dist: isWithinClrRange ? dis : null,
                        xbis: vv
                    };
                } else if (xbis == 0) {
                    var dis = isWithinClrRange && curpt ? mathAbs(relmousePt.x - curpt.x) : null;
                    return {
                        dist: dis,
                        xbis: xbis
                    };
                }
            });
        });
        var subseriesindex = [];
        var xpos = [];
        var min_xminimumBisected = xminimumBisected.map(function(d, i) {
            if (!d) {
                return null;
            }
            var arr = d.map(function(d) {
                return d.dist;
            });
            var min = d3_min(arr);
            var idx = arr.indexOf(min);
            subseriesindex[i] = min ? idx : null;
            xpos[i] = min ? d[idx].xbis : null;
            return min;
        });
        var seriespos = min_xminimumBisected.indexOf(d3_min(min_xminimumBisected));
        var itemPos = xpos[seriespos];
        // console.log("minbisect")
        // console.log(itemPos)
        return {
            itemPos: itemPos,
            seriesPos: seriespos,
            subseriesindex: subseriesindex[seriespos]
        };
    }

    function getClosestPoint(relmousePt, xscale, yscale) {
        var mindist_trigger,
            dist,
            is_within_bubble,
            bubbleRadius,
            cx,
            cy,
            prevBubbleInfo,
            series,
            subseries,
            item,
            seriesdata = chartObj.seriesdata,
            poly = NULL,
            prev_trigger_offset = NULL;

        var X = 'x',
            Y = 'y';
        if (axisrotated) {
            X = 'y';
            Y = 'x';
        }

        seriesdata.map(function(d, i) {
            var chartType = chartObj.chartTypes[i],
                type = chartType.value,
                chartName = chartType.name,
                isRangeChart = type == charttype.arearange,
                cur_trigger_offset = getTriggeringOffset(chartObj, type),
                plotOptions = d.plotoptions,
                commonPlotOptions = systemConf.chart.plot.plotoptions[chartName],
                markerOptions =
                    (plotOptions ? mergeJSON(commonPlotOptions.marker, plotOptions.marker, true) : commonPlotOptions.marker) || {},
                defaultMarkerSize = (markerOptions.enabled ? markerOptions.size || 1 : 1) * 10;

            if (!eventDisabled(d)) {
                if (chartObj.dataObject.isCombinationalChart) {
                    if (
                        !defined(mindist_trigger) ||
                        (defined(series)
                            ? cur_trigger_offset != NULL && cur_trigger_offset <= mindist_trigger
                            : cur_trigger_offset < prev_trigger_offset || prev_trigger_offset == NULL || cur_trigger_offset == NULL)
                    ) {
                        mindist_trigger = cur_trigger_offset;
                    }
                } else {
                    mindist_trigger = mindist_trigger == NULL ? cur_trigger_offset : mindist_trigger;
                }
            }

            var renderer = chartObj.renderer[i],
                yscindex = getParsedYAxisOrder(d.yaxiscolumnorder),
                ysc = yscale[yscindex].scale,
                hasNegValue = chartObj.dataObject.containsNegativeData(yscindex),
                prevData,
                yrangePos = isRangeChart ? 1 : NULL,
                polygonPoints = {};

            if (eventDisabled(d)) {
                return [NULL];
            } else {
                return d.data.map(function(data, j) {
                    data = renderer.seriesdataAfterUpdate ? renderer.seriesdataAfterUpdate[j] : data;
                    var cusPoints = isPolarAxisCategory
                        ? data.map(function(d, k) {
                              var x = chartObj.dataObject.getXYvalue(axisrotated, stacked(type), xscale, ysc, yscindex, d, 'x'),
                                  y = chartObj.dataObject.getXYvalue(axisrotated, stacked(type), xscale, ysc, yscindex, d, 'y');
                              return {
                                  x: x,
                                  y: y
                              };
                          })
                        : NULL;

                    return data.map(function(d, k) {
                        var isWithinClrRange = isWithinColorRange(dataset.getClr(d), chartObj),
                            condition;

                        if (isNullData(d, dataset, yscindex) || !isWithinClrRange) {
                            prevData = d;
                            return;
                        }

                        if (isPointChartType(type) || type === charttype.line) {
                            var cx = chartObj.dataObject.getXYvalue(axisrotated, stacked(type), xscale, ysc, yscindex, d, 'x'),
                                cy = chartObj.dataObject.getXYvalue(
                                    axisrotated,
                                    stacked(type),
                                    xscale,
                                    ysc,
                                    yscindex,
                                    d,
                                    'y',
                                    NULL,
                                    yrangePos
                                ),
                                shapeInfo = { cx: cx, cy: cy },
                                symbolbbox = renderer.getSymbolBBox ? renderer.getSymbolBBox(d) : NULL,
                                is_within_current_bubble;

                            if (symbolbbox != NULL) {
                                var symbolSegment = constructSegmentAroundDataPoint(
                                    d,
                                    dataset,
                                    xscale,
                                    ysc,
                                    yscindex,
                                    type,
                                    yrangePos,
                                    X,
                                    Y,
                                    symbolbbox
                                );
                                dist = distToSegment(relmousePt, symbolSegment[0], symbolSegment[1]);

                                var width = symbolbbox.width,
                                    height = symbolbbox.height,
                                    x = cx - width / 2,
                                    y = cy - height / 2;

                                shapeInfo.area = geometryUtils.areaOfRect(width, height);
                                is_within_current_bubble = geometryUtils.pointInRectangle(relmousePt, {
                                    x: x,
                                    y: y,
                                    width: width,
                                    height: height
                                });
                            } else {
                                dist = getMinDistByPtToPt(d, relmousePt, xscale, ysc, yscindex, type, yrangePos, X, Y);
                                var r = isBubbleChartType(type)
                                    ? chartObj.dataObject.getXYvalue(
                                          axisrotated,
                                          stacked(type),
                                          xscale,
                                          ysc,
                                          yscindex,
                                          d,
                                          'z',
                                          chartObj.zscale
                                      )
                                    : defaultMarkerSize;
                                r /= 2;

                                shapeInfo.radius = r; // TODO: remove
                                shapeInfo.area = geometryUtils.areaOfCircle(r);
                                is_within_current_bubble = dist <= r;

                                /*
                                 * If mouse point is inside the bubbles, find the bubble with minimum dist (from mouse point to centre)
                                 * Else find the bubble with minimum dist (from mouse point to arc)
                                 */
                                if (!is_within_current_bubble) {
                                    dist = dist - r; // distance to the arc
                                }
                            }

                            shapeInfo.dist = dist;
                            is_within_bubble = is_within_bubble || is_within_current_bubble;

                            if (type === charttype.bubblepie) {
                                var curpt = isNullData(d, dataset, yscindex)
                                    ? {
                                          x: 0,
                                          y: 0
                                      }
                                    : {
                                          x: cx,
                                          y: cy
                                      };
                                dist = pointInPie(relmousePt, curpt, i, j, d) ? dist : UNDEFINED;
                            }

                            if (is_within_bubble) {
                                if (is_within_current_bubble) {
                                    if (prevBubbleInfo) {
                                        var prevSurface = prevBubbleInfo.area,
                                            curSurface = shapeInfo.area,
                                            distanceBetweenCircles = mathSQRT(sqr(prevBubbleInfo.cx - cx) + sqr(prevBubbleInfo.cy - cy)),
                                            inside =
                                                isBubbleChartType(type) &&
                                                distanceBetweenCircles * 0.9 <= mathAbs(prevBubbleInfo.radius - shapeInfo.radius);

                                        condition =
                                            prevBubbleInfo.dist == NULL ||
                                            dist < prevBubbleInfo.dist ||
                                            !(prevSurface < curSurface && inside);
                                    } else {
                                        condition = true;
                                    }
                                }
                            } else {
                                condition = mindist_trigger == NULL || dist <= mindist_trigger;
                            }
                        } else {
                            var dist_n_polyPoints = getMinDistByPtToSegment(
                                d,
                                i,
                                j,
                                k,
                                data,
                                relmousePt,
                                prevData,
                                type,
                                xscale,
                                ysc,
                                yscindex,
                                hasNegValue,
                                isRangeChart,
                                cusPoints
                            );
                            dist = dist_n_polyPoints.distance;
                            polygonPoints = dist_n_polyPoints.polygonPoints;
                            prevData = d;

                            condition = mindist_trigger == NULL || dist <= mindist_trigger;
                        }

                        if (condition && dist != NULL) {
                            mindist_trigger = dist;
                            series = i;
                            subseries = j;
                            item = k;
                            poly = polygonPoints.points;
                            if (is_within_current_bubble) {
                                prevBubbleInfo = shapeInfo;
                            }
                        }
                    });
                });
            }
            prev_trigger_offset = cur_trigger_offset;
        });

        var customTriggerMinDistance = NULL;
        if (defined(series)) {
            customTriggerMinDistance = mindist_trigger;
        }

        return {
            seriesPos: series,
            itemPos: item,
            subseriesindex: subseries,
            poly: poly,
            customPoint: CustomPoints_th ? CustomPoints_th.getClosest(relmousePt, customTriggerMinDistance) : null
        };
    }

    /**
     * Hides all highlighter points.
     */
    function hideHighlighterPoints() {
        wrap.selectAll('#zcscatterhighlight,#zcbarhighlight,#zcareahighlight').style('display', NONE);
    }

    events_plot.hideHighlighterPoints = hideHighlighterPoints;

    function isNullData(data, dataset, yscindex) {
        if (chartObj.seriesdata[0].type == charttype.stkdarea) {
            return false;
        }
        return (
            data == null || isCatNull(dataset, 'x')(dataset.getX(data)) || isCatNull(dataset, 'y', yscindex)(dataset.getY(data, yscindex))
        );
    }

    function getMinDistByPtToSegment(
        d,
        i,
        j,
        k,
        data,
        relmousePt,
        prevData,
        type,
        xscale,
        ysc,
        yscindex,
        hasNegValue,
        isRangeChart,
        cusPoints
    ) {
        var dist = null;
        var polygonPoints = getPolygonPoints(
            i,
            j,
            k,
            data,
            prevData,
            data[k],
            data[k + 1],
            chartObj.dataObject,
            dataset,
            axisrotated,
            stacked(type),
            xscale,
            ysc,
            yscindex,
            hasNegValue,
            isRangeChart
        );

        var hasSinglePolygonPoint = polygonPoints.points.length == 1;
        if (isPolarAxisCategory) {
            polygonPoints.points = cusPoints;
        }

        var isWithinPoly = isAreaChartType(type) || isPolarAxisCategory ? pointInPolygon(relmousePt, polygonPoints.points) : true;
        if ((data.length == 1 || hasSinglePolygonPoint) && !isNullData(d, dataset, yscindex)) {
            // if series had one point
            dist = dist2(relmousePt, polygonPoints.curpt);

            // #ZC800
            var isComboChart = chartObj.dataObject.isCombinationalChart,
                areaSeriesCount = chartObj.dataObject.noofAreaEnabledSeries,
                triggerOffset = getTriggeringOffset(chartObj, charttype.line);

            if ((areaSeriesCount > 1 || (areaSeriesCount === 1 && isComboChart)) && dist > triggerOffset) {
                dist = null;
            }
        }
        if (isWithinPoly && !isNullData(d, dataset, yscindex) && !(data.length == 1 || hasSinglePolygonPoint)) {
            var pointA = distToSegment(relmousePt, polygonPoints.curpt, polygonPoints.curNext),
                pointB = distToSegment(relmousePt, polygonPoints.prevCur, polygonPoints.curpt),
                prevDataNull = k !== 0 && isNullData(prevData, dataset, yscindex);
            dist = k == 0 ? pointA : k == data.length - 1 ? pointB : prevDataNull ? pointA : d3_min([pointB, pointA]);

            // dist = (k == 0) ? distToSegment(relmousePt , polygonPoints.curpt , polygonPoints.curNext) :
            // 	(k == data.length-1) ? distToSegment(relmousePt, polygonPoints.prevCur, polygonPoints.curpt) :
            // 		d3_min([distToSegment(relmousePt, polygonPoints.prevCur, polygonPoints.curpt) , distToSegment(relmousePt, polygonPoints.curpt, polygonPoints.curNext)]);
        }

        return {
            distance: dist,
            polygonPoints: polygonPoints
        };
    }

    function getMinDistByPtToPt(d, relmousePt, xscale, ysc, yscindex, type, yrangePos, X, Y) {
        var curpt = {};
        curpt[X] = chartObj.dataObject.getXYvalue(axisrotated, stacked(type), xscale, ysc, yscindex, d, 'x');
        curpt[Y] = chartObj.dataObject.getXYvalue(axisrotated, stacked(type), xscale, ysc, yscindex, d, 'y', null, yrangePos);
        var dx = curpt.x - relmousePt[X];
        var dy = curpt.y - relmousePt[Y];
        return mathSQRT(sqr(dx) + sqr(dy));
    }

    function constructSegmentAroundDataPoint(d, dataset, xscale, ysc, yscindex, type, yrangePos, X, Y, rect) {
        var segments = [];
        if (isNullData(d, dataset, yscindex)) {
            return 0;
        } else {
            var curpt = {};
            curpt.x = chartObj.dataObject.getXYvalue(axisrotated, stacked(type), xscale, ysc, yscindex, d, 'x');
            curpt.y = chartObj.dataObject.getXYvalue(axisrotated, stacked(type), xscale, ysc, yscindex, d, 'y', null, yrangePos);

            if (axisrotated) {
                segments.push(
                    {
                        x: curpt.x,
                        y: curpt.y - rect.width / 2
                    },
                    {
                        x: curpt.x,
                        y: curpt.y + rect.width / 2
                    }
                );
            } else {
                segments.push(
                    {
                        x: curpt.x - rect.width / 2,
                        y: curpt.y
                    },
                    {
                        x: curpt.x + rect.width / 2,
                        y: curpt.y
                    }
                );
            }

            return segments;
        }
    }

    function getPolygonPoints(
        seriespos,
        subseriespos,
        itempos,
        data,
        prevData,
        curData,
        nextData,
        dataObject,
        dataset,
        axisrotated,
        stacked,
        xscale,
        ysc,
        yscindex,
        hasNegValue,
        isRangeChart
    ) {
        var prevpt, curpt, nextpt, prevBasept, curBasept, nextBasept;
        var yrangePos = isRangeChart ? 1 : null;

        curpt = isNullData(curData, dataset, yscindex)
            ? {
                  x: 0,
                  y: 0
              }
            : {
                  x: dataObject.getXYvalue(axisrotated, stacked, xscale, ysc, yscindex, curData, 'x'),
                  y: dataObject.getXYvalue(axisrotated, stacked, xscale, ysc, yscindex, curData, 'y', null, yrangePos)
              };
        prevpt =
            itempos == 0
                ? {
                      x: 0,
                      y: 0
                  }
                : isNullData(prevData, dataset, yscindex)
                ? curpt
                : {
                      x: dataObject.getXYvalue(axisrotated, stacked, xscale, ysc, yscindex, prevData, 'x'),
                      y: dataObject.getXYvalue(axisrotated, stacked, xscale, ysc, yscindex, prevData, 'y', null, yrangePos)
                  };
        nextpt =
            itempos == data.length - 1
                ? {
                      x: 0,
                      y: 0
                  }
                : isNullData(nextData, dataset, yscindex)
                ? curpt
                : {
                      x: dataObject.getXYvalue(axisrotated, stacked, xscale, ysc, yscindex, nextData, 'x'),
                      y: dataObject.getXYvalue(axisrotated, stacked, xscale, ysc, yscindex, nextData, 'y', null, yrangePos)
                  };

        curBasept = isNullData(curData, dataset, yscindex)
            ? {
                  x: 0,
                  y: 0
              }
            : addBasePointForArea(
                  dataObject,
                  axisrotated,
                  stacked,
                  seriespos,
                  subseriespos,
                  itempos,
                  xscale,
                  ysc,
                  yscindex,
                  hasNegValue,
                  curData,
                  isRangeChart
              );

        prevBasept =
            itempos == 0
                ? {
                      x: 0,
                      y: 0
                  }
                : isNullData(prevData, dataset, yscindex)
                ? curBasept
                : addBasePointForArea(
                      dataObject,
                      axisrotated,
                      stacked,
                      seriespos,
                      subseriespos,
                      itempos - 1,
                      xscale,
                      ysc,
                      yscindex,
                      hasNegValue,
                      prevData,
                      isRangeChart
                  );

        nextBasept =
            itempos == data.length - 1
                ? {
                      x: 0,
                      y: 0
                  }
                : isNullData(nextData, dataset, yscindex)
                ? curBasept
                : addBasePointForArea(
                      dataObject,
                      axisrotated,
                      stacked,
                      seriespos,
                      subseriespos,
                      itempos + 1,
                      xscale,
                      ysc,
                      yscindex,
                      hasNegValue,
                      nextData,
                      isRangeChart
                  );

        var prevCur =
            itempos == 0
                ? {
                      x: 0,
                      y: 0
                  }
                : isNullData(prevData, dataset, yscindex)
                ? curpt
                : {
                      x: (prevpt.x + curpt.x) / 2,
                      y: (prevpt.y + curpt.y) / 2
                  };
        var curNext =
            itempos == data.length - 1
                ? {
                      x: 0,
                      y: 0
                  }
                : isNullData(nextData, dataset, yscindex)
                ? curpt
                : {
                      x: (nextpt.x + curpt.x) / 2,
                      y: (nextpt.y + curpt.y) / 2
                  };

        var prevbaseCur =
            itempos == 0
                ? {
                      x: 0,
                      y: 0
                  }
                : isNullData(prevData, dataset, yscindex)
                ? curBasept
                : {
                      x: (prevBasept.x + curBasept.x) / 2,
                      y: (prevBasept.y + curBasept.y) / 2
                  };
        var curbaseNext =
            itempos == data.length - 1
                ? {
                      x: 0,
                      y: 0
                  }
                : isNullData(nextData, dataset, yscindex)
                ? curBasept
                : {
                      x: (nextBasept.x + curBasept.x) / 2,
                      y: (nextBasept.y + curBasept.y) / 2
                  };

        var points = [];

        var temp =
            isNullData(prevData, dataset, yscindex) && isNullData(nextData, dataset, yscindex)
                ? points.push(curpt)
                : itempos == 0 || isNullData(prevData, dataset, yscindex)
                ? points.push(curpt, curNext, curbaseNext, curBasept)
                : itempos == data.length - 1 || isNullData(nextData, dataset, yscindex)
                ? points.push(prevCur, curpt, curBasept, prevbaseCur)
                : points.push(prevCur, curpt, curNext, curbaseNext, curBasept, prevbaseCur);
        return {
            points: points,
            prevCur: prevCur,
            curNext: curNext,
            curpt: curpt
        };
    }

    function addBasePointForArea(
        dataObject,
        axisrotated,
        stacked,
        seriespos,
        subseriespos,
        itemPos,
        xsc,
        ysc,
        yscindex,
        hasNegValue,
        point,
        isRangeChart
    ) {
        var basePoint;
        var yrangePos = isRangeChart ? 0 : null;
        var temp;
        if (axisrotated) {
            var xval = dataObject.getXYvalue(axisrotated, stacked, xsc, ysc, yscindex, point, 'y', null, yrangePos);
            var yval = isRangeChart
                ? dataObject.getXYvalue(axisrotated, stacked, xsc, ysc, yscindex, point, 'x', null, yrangePos)
                : ysc.range()[0];
            var accessor = isRangeChart && dataset.getY(point, yscindex, 0) < 0 ? mathMin : mathMax;
            if (stacked) {
                temp = seriespos == 0 && subseriespos == 0 && itemPos == 0 ? mathMax(ysc.range()[0], ysc(hasNegValue)) : ysc(point.y0);
                basePoint = {
                    x: temp,
                    y: xval
                };
            } else {
                temp = ysc.bandwidth != null ? ysc.origRange[0] + ysc.bandwidth() : accessor(yval, ysc(hasNegValue));
                basePoint = {
                    x: temp,
                    y: xval
                };
            }
        } else {
            var xval = dataObject.getXYvalue(axisrotated, stacked, xsc, ysc, yscindex, point, 'x');
            var yval = isRangeChart
                ? dataObject.getXYvalue(axisrotated, stacked, xsc, ysc, yscindex, point, 'y', null, yrangePos)
                : ysc.range()[0];
            var accessor = isRangeChart && dataset.getY(point, yscindex, 0) < 0 ? mathMax : mathMin; //#ZC1336
            if (stacked) {
                temp = seriespos == 0 && subseriespos == 0 && itemPos == 0 ? mathMin(ysc.range()[0], ysc(hasNegValue)) : ysc(point.y0);
                basePoint = {
                    x: xval,
                    y: temp
                };
            } else {
                temp = ysc.bandwidth != null ? yval + ysc.bandwidth() : accessor(yval, ysc(hasNegValue));
                basePoint = {
                    x: xval,
                    y: temp
                };
            }
        }
        return basePoint;
    }

    function hideUnselectedWrap(wrap) {
        wrap.style('display', NONE);
    }

    function isWithin(dataPt, mousePt, reversed) {
        var x1 = dataPt.x;
        var y1 = dataPt.y;
        var x2 = x1 + dataPt.width;
        var y2 = y1 + dataPt.height;
        if (reversed) {
            axisrotated ? (x2 = (x1 = x1 - dataPt.width) + dataPt.width) : (y2 = y1 - dataPt.height);
        }

        var condition =
            x1 <= mousePt.x &&
            mousePt.x <= x2 &&
            (reversed && !axisrotated ? y2 <= mousePt.y && mousePt.y <= y1 : y1 <= mousePt.y && mousePt.y <= y2);
        if (condition) {
            return 0;
        } else {
            return null;
        }
    }

    function isWithinPlotarea(xval, yval) {
        return isWithin(
            {
                x: plotarea.left + (chartObj.scrolledX || 0),
                y: plotarea.top + (chartObj.scrolledY || 0),
                width: plotarea.width,
                height: plotarea.height
            },
            {
                x: plotarea.left + xval,
                y: plotarea.top + yval
            }
        );
    }

    function sqr(x) {
        return x * x;
    }

    function dist2(v, w) {
        return sqr(v.x - w.x) + sqr(v.y - w.y);
    }

    function distToSegmentSquared(p, v, w) {
        var l2 = dist2(v, w);
        if (l2 === 0) {
            return dist2(p, v);
        }
        var t = ((p.x - v.x) * (w.x - v.x) + (p.y - v.y) * (w.y - v.y)) / l2;
        if (t < 0) {
            return dist2(p, v);
        }
        if (t > 1) {
            return dist2(p, w);
        }
        return dist2(p, {
            x: v.x + t * (w.x - v.x),
            y: v.y + t * (w.y - v.y)
        });
    }

    function distToSegment(p, v, w) {
        return mathSQRT(distToSegmentSquared(p, v, w));
    }

    function lineDistance(point1, point2) {
        var xs = 0;
        var ys = 0;

        xs = point2.x - point1.x;
        xs = xs * xs;

        ys = point2.y - point1.y;
        ys = ys * ys;

        return mathSQRT(xs + ys);
    }

    function isAbove(a, b, c) {
        //a = line point 1; b = line point 2; c = point to check against.
        return (b.x - a.x) * (c.y - a.y) - (b.y - a.y) * (c.x - a.x) > 0 && (b.x - a.x) * (c.y - b.y) - (b.y - a.y) * (c.x - b.x) > 0;
        // var numerator = ((b.y - a.y)*c.x) - ((b.x - a.x)*c.y) + (b.x*a.y) - (b.y*a.x);
        // return numerator > 0
    }

    function pointInPolygon(point, poly) {
        // ray-casting algorithm based on
        // www.ecse.rpi.edu/Homepages/wrf/Research/Short_Notes/pnpoly.html
        var xi,
            xj,
            yi,
            yj,
            i,
            intersect,
            x = point.x,
            y = point.y,
            inside = false;
        for (var i = 0, j = poly.length - 1; i < poly.length; j = i++) {
            (xi = poly[i].x),
                (yi = poly[i].y),
                (xj = poly[j].x),
                (yj = poly[j].y),
                (intersect = yi > y != yj > y && x < ((xj - xi) * (y - yi)) / (yj - yi) + xi);
            if (intersect) {
                inside = !inside;
            }
        }
        return inside;
    }

    function pointInPie(point, curpt, seriespos, subseriespos, d) {
        var pieAngle = getBubblePieAngle(chartObj, seriespos, subseriespos, d);

        if (!pieAngle) {
            return false;
        }

        var angle = mathAtan2(point.x - curpt.x, -(point.y - curpt.y));
        return isAngleInside(pieAngle.start, pieAngle.end, angle);
    }

    function eventDisabled(d) {
        /*
         * returns true if the series is filtered or event is disabled for that series
         */
        var plotOptions = d.plotoptions;
        var cName = getChartType(d.type || chartObj.userdata.seriesdata.type, 'string');
        var commonPlotOptions = systemConf.chart.plot.plotoptions[cName];
        return d.disabled || validateNullVal(getEventConf(plotOptions, commonPlotOptions, chartObj, 'enabled'), true) == false;
    }

    function getCurrentDataPoint(seriesIndex, subseriesIndex, itemIndex, itemPosShared, parent) {
        var renderer = chartObj.renderer[seriesIndex],
            currentSeriesdata =
                renderer && renderer.seriesdataAfterUpdate
                    ? renderer.seriesdataAfterUpdate[subseriesIndex]
                    : chartObj.seriesdata[seriesIndex].data[subseriesIndex],
            xPos = tooltipShared && defined(itemPosShared) ? itemPosShared[seriesIndex][subseriesIndex][0] : itemIndex; // TODO: Events_55.json duplicate x values are available. Have to work on how to show all in the tooltip and marker highlight. itemPosShared[seriesIndex][subseriesIndex] is an array which contains all the xpos for the current subseries.

        return parent ? currentSeriesdata : currentSeriesdata[xPos];
    }

    /**
     *
     * @param seriesIndex
     * @param subSeriesIndex
     * @param itemIndex
     * @return {{x, y, width, height}}
     */
    function getPointRectangle(seriesIndex, subSeriesIndex, itemIndex) {
        if (!defined(seriesIndex) || !defined(itemIndex)) {
            return;
        }

        subSeriesIndex = pick(subSeriesIndex, 0);

        var seriesRenderer = chartObj.renderer[seriesIndex],
            currentPoint = seriesRenderer.seriesdataAfterUpdate[subSeriesIndex][itemIndex],
            barOptions = seriesRenderer.commonOption || getBarCommonOption(chartObj.rendererConf[seriesIndex], seriesIndex);

        return getBarPosition(chartObj, currentPoint, subSeriesIndex, barOptions);
    }

    function getRect(highlightpoint, radius, yaxisReversed) {
        if (isHeatMapChartType(chartObj.chartTypes[0].value)) {
            var commonPlotOptions = chartObj.systemConf.chart.plot.plotoptions.heatmap;
            var symbolPath = commonPlotOptions.symbol && getSVGsymbol(commonPlotOptions.symbol);
            if (symbolPath) {
                return symbolPath(mathMin(highlightpoint.width, highlightpoint.height));
            } else {
                return roundedRect(
                    -highlightpoint.width / 2,
                    -highlightpoint.height / 2,
                    highlightpoint.width,
                    highlightpoint.height,
                    radius,
                    null,
                    axisrotated
                );
            }
        } else {
            return roundedRect(
                highlightpoint.x,
                highlightpoint.y,
                highlightpoint.width,
                highlightpoint.height,
                radius,
                yaxisReversed,
                axisrotated
            );
        }
    }

    function getRectTranslation(highlightpoint) {
        var tx = plotarea.left,
            ty = plotarea.top;
        if (isHeatMapChartType(chartObj.chartTypes[0].value)) {
            tx += highlightpoint.x + highlightpoint.width / 2;
            ty += highlightpoint.y + highlightpoint.height / 2;
        }
        return [tx, ty];
    }

    function parseHierarchyLevel(level) {
        level = level.split('_');
        return {
            levelIndices: level.map(Number),
            seriesIndex: +level[0],
            itemIndex: +level[level.length - 1]
        };
    }
    function getAvailableBubblePieKeys() {
        var bubblePieData = chartObj.dataObject.bubblepieData;
        return (
            bubblePieData &&
            d3_merge(bubblePieData)
                .filter(function(item) {
                    //Bubblepie with missing categories(Events_55.json)
                    return item;
                })
                .map(function(item) {
                    var dt = item.reference;
                    return bubblepieUtils.getKey(NULL, NULL, NULL, [item.x, item.y]); //#ZC1323
                })
        );
    }

    /**
     * Checks if a user has allowed the fault function
     *
     * @param userFunction
     * @param event
     * @return {*|boolean}
     */
    function isDefaultAllowed(userFunction, event) {
        return !defined(userFunction) || (userFunction && event.allowDefault === true);
    }

    function triggerUserCallback(callback, point, event, extraParams) {
        var userArguments;

        // get arguments for threshold marker
        if (point.type === 'threshold-marker') {
            var key = point.thresholdType === 'x' ? 'x' : `y.groups[${point.yColumnIndex}]`;
            userArguments = getPropVal(chartObj.axes, `${key}.threshold.line`).getMetaInfo(point);
        } else {
            userArguments = getUserArguments(point);
        }

        // set extra params
        if (defined(extraParams) && typeof extraParams === 'object') {
            userArguments = mergeJSON(userArguments, extraParams, true, true);
        }
        // fire the event
        invokeFunction(callback, event, userArguments, chartObj);
    }

    /**
     *
     * @param point
     * @return {*}
     */
    function getUserArguments(point) {
        var params,
            rangePoint = point.currentQualitativeRangePoint,
            dataPoint = pick(point.point, point.data) || {},
            levels = pick(point.level, dataPoint && dataPoint.level);

        // it is a hierarchy chart.
        // indexes need to be parsed
        if (defined(levels)) {
            params = parseHierarchyLevel(levels);
        } else {
            params = {
                seriesIndex: pick(point.seriesPos, dataPoint.seriesIndex),
                subSeriesIndex: pick(point.subseriesindex, point.subseriesPos, dataPoint.subSeriesIndex),
                itemIndex: pick(point.itemPos, dataPoint.itemIndex) //TODO: priority should be corrected
            };
        }

        if (isArray(point.subseriesindex)) {
            point.subseriesindex.some(function(value) {
                if (defined(value)) {
                    return (params.subSeriesIndex = value);
                }
            });
        }

        // handle bullet range points
        if (defined(rangePoint)) {
            params.rangeIndex = rangePoint.rangeIndex;
            params.targetIndex = rangePoint.targetIndex;
        }

        if (defined(params.seriesIndex) && defined(params.itemIndex)) {
            if (defined(dataPoint) && isArray(dataPoint.data)) {
                dataPoint = dataPoint.data;
            }

            if (!isArray(dataPoint)) {
                params.subSeriesIndex = pick(params.subSeriesIndex, 0);
                dataPoint = getCurrentDataPoint(params.seriesIndex, params.subSeriesIndex, params.itemIndex, point.itemPosShared);

                // since all indexes has been set, set the label index alone.
                if (defined(dataPoint.level)) {
                    params.levelIndices = parseHierarchyLevel(dataPoint.level).levelIndices;
                }
            }

            // TODO has to be revamp. Since we are passing all required params, It can be access through API.
            params.data = chartObj.dataset.getMetaInfoObj(chartObj, dataPoint, params.seriesIndex);
        }
        // pass the actual point
        params.point = dataPoint;

        return params;
    }

    // add helper function
    events_plot.helpers = {
        isDefaultAllowed: isDefaultAllowed,
        triggerUserCallback: triggerUserCallback,
        eventDisabled: eventDisabled
    };

    events_plot.hoverPolygon = hoverPolygon;
    events_plot.hoverScatter = hoverScatter;
    events_plot.getPolygonPoints = getPolygonPoints;

    return events_plot;
};

export var _plotEvents = events.mouseevents;
