// @zohocharts/zc/src/geo-core/map-series.js
//MapSeries.js
import { getPropVal, mergeJSON, pick } from '../utils/zcutils';
import { gradient_helpers } from '../effects/gradient-helpers';
import { getFillColor, getStrokeColor } from '../utils/colorutils';
import { GeoHighlighter } from "./geo-highlight";

/**
 *
 * @param {Object} series
 * @param {GeoMap} geoMap
 * @param {number} index
 * @constructor
 */
export function MapSeries(series, geoMap, index) {
    this.series = series;
    this.geoMap = geoMap;
    this.index = index;

    this.initialize();
}

/**
 *
 * @param geoMap
 * @param chartName
 * @return {*}
 */
MapSeries.getPlotOptions = function(geoMap, chartName) {
    return getPropVal(geoMap.systemConf, `chart.plot.plotoptions.${chartName}`);
};

MapSeries.prototype.initialize = function() {
    var mapSeries = this,
        geoMap = mapSeries.geoMap,
        chart = geoMap.chartTypes[mapSeries.index],
        gradientDimension = gradient_helpers.getDimension(geoMap, mapSeries.index);

    geoMap.context = pick(geoMap.geo.pointLayer, geoMap.geo.featureLayer);
    mapSeries.isCanvasMode = geoMap.geo.isCanvasMode;

    mapSeries.plotOptions = mergeJSON(MapSeries.getPlotOptions(geoMap, chart.name), pick(mapSeries.series.plotoptions, {}), true);
    mapSeries.fillColor = getFillColor(geoMap, mapSeries.index, null, gradientDimension, null, mapSeries.renderingMode);
    mapSeries.strokeColor = getStrokeColor(geoMap, mapSeries.index, gradientDimension);

    this.getHighlighter();
    return mapSeries;
};

/**
 *
 */
MapSeries.prototype.update = function() {
    this.render(true);
};

MapSeries.prototype.getHighlighter = function() {
    this.highlighter = new GeoHighlighter(this.geoMap, this);
};
