// @zohocharts/zc/src/geo-core/map-constructor.js
//map-constructor.js
import { Registry } from '../core/zc-registry';
import { VizLayout } from '../core/main-constructor';
import { inherit, defined, toCapitalize, mergeJSON, pick, getPropVal, each, arrayIncludes } from '../utils/zcutils';
import { MapLayer } from './map-layer';
import { $Browser } from '../core/browser-detect';
import { stateChangeListener } from '../events/eventActions';
import { plot, chartcategory, GeoSeriesTypes, zmaps, tooltip } from '../core/main-zc-init';
import { createElement } from '../utils/zcsvgutils';
// import { scale } from '../core/main-zc-init';
import { GeoLabels } from './map-labels';
import { MapDataProcessor } from './map-data-processor';
import { MapDataSet } from './map-data-parser';
import { d3_ease, topojson, d3_geoPath, d3_geoGraticule, d3_zc } from '../core/d3-constants';
// import { tooltip_block_helpers } from '../tooltip/tooltip-blockPosition';
import { maps } from '../core/chart-instance-manager';
import { GeoPlotRenderer } from './map-renderer';
import { d3_map } from '../core/d3-constants';
import { GeoUtils } from './map-utils';
// import { getPlotBoundingRect } from '../scroll/defaultScroll';
// import { setDefaultOptions } from '../core/defaulttheme';
// import { ZC_ChartEvents } from '../events/chart-events';
// import { initTooltip } from '../events/visualEffects/controllers/tooltip_c';
import { BehaviouralEvents } from '../behaviours/behavioural-events';
import { __getBaseProjectionTransform } from '../behaviours/visual-effects/map/tools/base-projection-transform';
import { ZoomToPosition } from '../behaviours/visual-effects/map/tools/zoom-to-position';
import { ZC_MapEvents } from './zc_map_events';

let ZoomD3 = Registry.getComponent('ZoomD3');
/**
 * The default map collection
 * @type {{}}
 */
export var mapCollections = {};

/**
 *
 * @constructor
 */
export function GeoMap(container, data) {
    // define axis object for color axis (heatmap|choropleth)
    this.axes = {
        x: {},
        y: {}
    };

    VizLayout.call(this, container, data);
    this.initialize();
}

inherit(GeoMap, VizLayout);
GeoMap.helpers = VizLayout.helpers; //IE

// add a global type holder to keep all series types.
GeoMap.SeriesTypes = GeoSeriesTypes;

/**
 *
 * @return {{name: string}}
 */
GeoMap.getDefaultProjection = function() {
    return {
        name: 'geoMercator'
    };
};

/**
 *
 * @param layers
 * @param layerName
 * @return {*}
 */
GeoMap.getGeoJSON = function(layers, layerName) {
    // if the layer is Topojson convert it to Geojson
    // because d3 tools works purely in Geojson
    if (defined(layers) && MapLayer.GEO_FORMATS.has(layers.type) && layers.type === 'Topology') {
        // get the default projection settings from data
        // because it will be whipped off while geojson conversion.
        var projectionSettings = layers.projection;

        // convert the topojson to geojson
        layers = topojson.feature(layers, layers.objects[layerName]);

        // apply the default projection settings in converted geojson
        if (defined(projectionSettings)) {
            layers.projection = projectionSettings;
        }
    }
    return layers;
};

/**
 * The main map initializer.
 */
GeoMap.prototype.initialize = function() {
    // inti the color axes
    var geoMap = this,
        userData = geoMap.userdata,
        userMapConfig = userData.map,
        hasSvgOnly = geoMap.chartTypes.some(function(chartType) {
            return arrayIncludes(chartcategory.svgOnly, chartType.value);
        });

    // add a ID
    geoMap.id = `ZC_GEO_${geoMap.idNumeric}`;

    if ($Browser.IE && parseFloat($Browser.VERSION) < 11) {
        geoMap.fixIE = true;
    }

    // setup the internal config container
    geoMap.geo = {
        layerCount: 0,
        extraLayers: {},
        extraLayersRenderingOrder: ['globe', 'graticule', 'allAreas', 'backgroundLayer'],
        utils: {},
        highlightUtils: {},
        rawLayers: [],
        customLayers: d3_map()
    };
    geoMap.timerObj = new Array(0);
    // process the data
    geoMap.processUserData(geoMap.themeoptions);

    // fill color error
    geoMap.plot = {};
    geoMap.geo.isCanvasMode = geoMap.systemConf.map.renderingMode.toLowerCase() !== 'svg';
    geoMap.geo.isPlotCanvasMode = geoMap.systemConf.chart.plot.renderer.mode.toLowerCase() !== 'svg' && !hasSvgOnly;

    // create all the needed elements
    GeoMap.helpers.constructChartGroups(geoMap);

    // get the series holder
    geoMap.seriesContainer = geoMap.wrap.select('.seriesholder');
    geoMap.chartGroup = geoMap.svg.select('.chartgroup');

    // start rendering
    let legendHighlight = Registry.getComponent('LegendHighlight');
    geoMap.legendHighlight = legendHighlight(this, this.wrap);

    geoMap.render();
};

/**
 *
 */
GeoMap.prototype.render = function() {
    var geoMap = this,
        systemConfig = geoMap.systemConf,
        animate = systemConfig.map.animation.enabled && !geoMap.updating;

    animate = animate && ((geoMap.dataObject.hasHeatmap && !geoMap.dataObject.hasPointChart) || !geoMap.seriesdata.length);

    /**
     * **************************************************************
     *
     * Setting up the main container and legend
     */
    // get the series holder
    geoMap.seriesContainer = geoMap.wrap.select('.seriesholder');
    geoMap.chartGroup = geoMap.svg.select('.chartgroup');

    // after processing set the chart dimension
    geoMap.setChartDimension();
    // render the container
    geoMap.updateCanvasArea();

    geoMap.applyChartProperties();

    // if all areas is false and data point also 0 then call no data
    if (!geoMap.hasGeoLayer || (!systemConfig.map.allAreas && !geoMap.dataObject.totalDataCount)) {
        // report no map layer found
        // #ZC1652
        geoMap.flushed = true;
        geoMap.clearMap();
        geoMap.hideLoading(geoMap.container);
        geoMap.handleNoData(systemConfig.noDataHandler);
        geoMap.chartRendered = true;
        geoMap.handleHiddenContainer(true);
        stateChangeListener(geoMap, systemConfig.canvas.events, 'chart', geoMap.updating ? 'onreload' : 'onload');
        return;
    } else if (!geoMap.dataObject.totalDataCount) {
        geoMap.flushed = true;
        geoMap.clearMap();
        geoMap.clearChart(null, null, true);
    }

    geoMap.clearNoData(systemConfig.noDataHandler);
    /**
     * **************************************************************
     *
     * Setting map rendering area, projection and path generator
     */

    // color dependency
    geoMap.getChartSeriesColor();
    geoMap.parseScale();

    if (geoMap.hasSeriesPoints) {
        // render the axis
        Registry.getComponent('legend.render')(geoMap);

        //#ZC1444
        tooltip.block_helpers && tooltip.block_helpers.updatePlotarea4BlockTooltip(geoMap);

        // fix the legend position based on plot area
        // geoMap.fixLegendPosition();
    }

    //Adding block level annotations
    let annotation = Registry.getComponent('annotation');
    if (!annotation._empty) {
        annotation.addBlockAnnotations(geoMap);
    }

    geoMap.updatePlotareaProp();

    // setup the main container layer (svg|canvas)
    geoMap.setupFeatureContainerLayer();

    // draw the default background layer
    geoMap.setUpMapLayer();

    // IE fix
    // if animation is enabled
    if (animate) {
        geoMap.animateSelection = geoMap.container
            .selectAll('.geomap-features-series,.geomap-features-extras,.xychart,.static,.label-container,.labelholder')
            .style('opacity', 0);
    }

    // fire plot start event on first load
    stateChangeListener(geoMap, systemConfig.map.events, 'map', 'init');

    // render the world out line
    geoMap.renderGlobe();

    // render graticule
    geoMap.renderGraticule();

    // render the map background layer
    geoMap.renderMapLayer();

    stateChangeListener(geoMap, systemConfig.map.events, 'map', geoMap.updating ? 'onreload' : 'onload');

    // render the custom layers added by the user
    if (geoMap.geo.customLayers.size()) {
        geoMap.geo.customLayers.each(layer => layer && layer.update());
    }

    // return
    if (geoMap.plot.renderer) {
        geoMap.plot.renderer.draw(true);
    } else if (geoMap.hasSeriesPoints) {
        geoMap.plot = {
            renderer: new GeoPlotRenderer(geoMap)
        };
        geoMap.plot.renderer.draw();
    }

    geoMap.drawPlotBorder();

    geoMap.renderLabels();

    geoMap.handleEvent();

    // ZC1414
    geoMap.drawPlotareaClippath();

    if (!geoMap.plot.renderer) {
        geoMap.animate();
        stateChangeListener(geoMap, systemConfig.canvas.events, 'chart', geoMap.updating ? 'onreload' : 'onload');
        geoMap.updating = false;
    }
    geoMap.flushed = false;
};

/**
 * this will create the main container element based on the rendering mode.
 */
GeoMap.prototype.setupFeatureContainerLayer = function() {
    var geoMap = this,
        geoConfig = geoMap.geo,
        layerNameCapitalized = toCapitalize(geoConfig.isCanvasMode ? 'canvas' : 'svg'),
        isGeoCanvasMode = geoConfig.isCanvasMode,
        isPlotCanvasMode = geoConfig.isPlotCanvasMode,
        mapArea = (geoConfig.mapArea = Object.create(geoMap.plotarea));

    // translate the series holder.
    geoMap.seriesContainer.attr('transform', `translate(${mapArea.left},${mapArea.top})`);

    if (isGeoCanvasMode) {
        geoConfig.featureLayerExtra = plot.canvasRenderer.updateCanvasContext(
            geoMap,
            'geomap-features-extras',
            '.static',
            geoMap.chartParentElement
        );
        geoConfig.geoLabelLayer = geoConfig.featureSeriesLayer = plot.canvasRenderer.updateCanvasContext(
            geoMap,
            'geomap-features-series',
            '.static',
            geoMap.chartParentElement
        );
    } else {
        // geo config
        geoConfig.featureSeriesLayer = createElement('g', '.feature-layer-series', [1], geoMap.seriesContainer);
        geoConfig.geoLabelLayer = createElement('g', '.geo-label-layer', [1], geoMap.seriesContainer);
    }

    if (isPlotCanvasMode) {
        // data point config
        geoConfig.pointLabelLayer = geoMap.labelHolderContext = plot.canvasRenderer.updateCanvasContext(
            geoMap,
            'label-container',
            geoMap.fixIE ? '.static' : 'div.legendarea',
            geoMap.chartParentElement
        );
        geoConfig.pointLayer = plot.canvasRenderer.updateCanvasContext(geoMap, 'zc-canvas', '.static', geoMap.chartParentElement);
    } else {
        // data point config
        geoConfig.labelLayer = geoMap.staticPaneSVG.select('.labelholder').attr('transform', `translate(${mapArea.left},${mapArea.top})`);
    }
};

/**
 *
 * @param layer
 */
GeoMap.prototype.setupLayerSettings = function(layer) {
    var geoMap = this,
        geoConfig = geoMap.geo,
        projection = geoConfig.utils.projection,
        rotation = projection && projection.rotate && projection.rotate(),
        behaviours = (geoMap.behaviouralEvents && geoMap.behaviouralEvents.eventManager.behaviours) || {},
        plotZoom = behaviours.PLOT_ZOOM;

    // setup the geo path renderer
    projection = geoConfig.utils.projection = geoConfig.highlightUtils.projection = geoMap.getProjection(layer);

    if (plotZoom) {
        let fitExtent = geoMap.getProjectionExtent();
        let transform = plotZoom.instance.getTransform(),
            baseProjectionTransform = __getBaseProjectionTransform(geoMap),
            center = ZoomD3.centroid(GeoUtils.geoBoundToRect(fitExtent)),
            translate = baseProjectionTransform.translate,
            scale = baseProjectionTransform.scale;

        if (geoMap.geo.isGlobeMap) {
            projection.scale(baseProjectionTransform.scale * transform.k);
            projection.translate(center);
            projection.rotate(rotation);
        } else {
            projection.scale(scale * transform.k);
            projection.translate([transform.x + translate[0] * transform.k, transform.y + translate[1] * transform.k]);
        }
    }

    // if (!defined(geoConfig.utils.projection)) {
    //     // projection = geoConfig.utils.projection = geoConfig.highlightUtils.projection = geoMap.getProjection(layer);
    // } else {
    //     let fitExtent = geoMap.getProjectionExtent();
    //
    //     if (geoMap.plotZoom) {
    //         let transform = geoMap.plotZoom.getTransform(),
    //         baseProjectionTransform = geoMap.plotZoom.__getBaseProjectionTransform(),
    //         center = Zoom.centroid(GeoUtils.geoBoundToRect(fitExtent)),
    //         translate = baseProjectionTransform.translate,
    //         scale = baseProjectionTransform.scale;
    //
    //         // if (geoMap.geo.isGlobeMap) {
    //         //     projection.scale(baseProjectionTransform.scale * transform.k);
    //         //     projection.translate(center);
    //         // } else {
    //         //     projection.scale(scale * transform.k);
    //         //     projection.translate([transform.x + translate[0] * transform.k, transform.y + translate[1] * transform.k]);
    //         // }
    //     } else {
    //         // projection.fitExtent(fitExtent, layer);
    //     }
    // }

    // set up the geo path with map projection
    geoConfig.utils.geoPath = d3_geoPath().projection(projection);
    geoConfig.highlightUtils.geoPath = d3_geoPath().projection(projection);

    // if canvas mode set the context in the geo path.
    if (geoConfig.isCanvasMode) {
        geoConfig.utils.geoPath.context(geoConfig.featureSeriesLayer);
    }
};

/**
 *
 * @param layer
 */
GeoMap.prototype.getProjection = function(layer) {
    var geoMap = this,
        geoConfig = geoMap.geo,
        defaultLayer = geoConfig.defaultLayer.layerGeojson,
        userConfig = geoMap.systemConf.map,
        projectionSettings = pick(userConfig.projection, layer.projection, defaultLayer.projection),
        projection,
        geojsonLayer = layer,
        fitExtent = geoMap.getProjectionExtent();

    geoMap.geo.isGlobeMap = projectionSettings && projectionSettings.name === 'geoOrthographic';

    if (!defined(projectionSettings)) {
        projectionSettings = GeoMap.getDefaultProjection();
    } else if (geoMap.geo.isGlobeMap) {
        // if it is a globe map, render the complete globe
        geojsonLayer = {
            type: 'Sphere'
        };
    }
    // setup the geo path renderer
    projection = d3_zc[projectionSettings.name]();

    for (var prop in projectionSettings) {
        if (projectionSettings.hasOwnProperty(prop) && prop !== 'name' && projection[prop]) {
            projection[prop](projectionSettings[prop]);
        }
    }

    projection.fitExtent(fitExtent, geojsonLayer);

    return projection;
};

GeoMap.prototype.getProjectionExtent = function() {
    var geoMap = this,
        mapArea = geoMap.geo.mapArea,
        userConfig = geoMap.systemConf.map,
        extent = userConfig.extent;

    return [[extent.left, extent.top], [mapArea.width - extent.right, mapArea.height - extent.bottom]];
};

/**
 *
 * @param layer
 * @param styles
 * @param name
 * @param type
 * @param useMesh
 * @return {MapLayer}
 */
GeoMap.prototype.addLayer = function(layer, styles, name, type, useMesh) {
    var geoMap = this,
        geoConfig = geoMap.geo,
        layerId = geoConfig.layerCount++,
        types = {
            series: geoConfig.featureSeriesLayer,
            extra: geoConfig.featureLayerExtra,
            highlight: geoConfig.highlightLayer
        },
        layerContainer = types[type];

    name = name || `map-layer-${layerId}`;

    if (!geoConfig.isCanvasMode) {
        layerContainer = createElement('g', `.${name}`, [name], geoConfig.featureSeriesLayer);
    }

    // layer, util, layerContainer, styles
    return new MapLayer(layer, geoConfig.utils, layerContainer, styles, geoMap, useMesh);
};

/**
 *
 * @param layer
 * @param styles
 */
GeoMap.prototype.addGeoLayer = function(layer, styles) {
    var geoMap = this,
        customLayers = geoMap.geo.customLayers,
        layerId = `custom-geo-layer-${geoMap.geo.layerCount++}`,
        mapLayer = geoMap.addLayer(layer, styles, layerId, 'extra', false);

    // save it
    customLayers.set(layerId, mapLayer);

    return layerId;
};

/**
 *
 * @param layerId
 */
GeoMap.prototype.removeGeoLayer = function(layerId) {
    var geoMap = this;

    if (geoMap.geo.customLayers.has(layerId)) {
        geoMap.geo.customLayers.get(layerId).destroy();
        delete geoMap.geo.customLayers.remove(layerId);
    }
    GeoUtils.clearContext(geoMap.geo.featureLayerExtra);
    geoMap.repaintGeoLayer();
};

/**
 *
 */
GeoMap.prototype.setupSvgLayer = function(styles) {
    var geoMap = this,
        geoConfig = geoMap.geo,
        mapArea = geoConfig.mapArea,
        userConfig = geoMap.systemConf.map;

    // move the feature layer
    geoConfig.featureSeriesLayer = createElement('g', '.feature-layer-series', [1], geoMap.seriesContainer);

    geoMap.seriesContainer.attr('transform', `translate(${mapArea.left},${mapArea.top})`);

    geoConfig.labelLayer = geoMap.staticPaneSVG.select('.labelholder').attr('transform', `translate(${mapArea.left},${mapArea.top})`);

    // plot labels rely on labelHolderContext
    if (geoMap.systemConf.chart.plot.renderer.mode === 'canvas') {
        geoMap.labelHolderContext = plot.canvasRenderer.updateCanvasContext(
            geoMap,
            'label-container',
            geoMap.fixIE ? '.static' : 'div.legendarea',
            geoMap.chartParentElement
        );
        geoConfig.pointLayer = plot.canvasRenderer.updateCanvasContext(geoMap, 'zc-canvas', '.static', geoMap.chartParentElement);
    }
};

/**
 *
 */
GeoMap.prototype.setupCanvasLayer = function() {
    var geoMap = this,
        geoConfig = geoMap.geo;

    // insert before static svg
    geoConfig.featureLayerExtra = plot.canvasRenderer.updateCanvasContext(
        geoMap,
        'geomap-features-extras',
        '.static',
        geoMap.chartParentElement
    );
    geoConfig.featureSeriesLayer = plot.canvasRenderer.updateCanvasContext(
        geoMap,
        'geomap-features-series',
        '.static',
        geoMap.chartParentElement
    );

    if (geoMap.dataObject.hasPointChart) {
        geoConfig.pointLayer = plot.canvasRenderer.updateCanvasContext(geoMap, 'zc-canvas', '.static', geoMap.chartParentElement);
    }

    geoConfig.labelLayer = geoMap.labelHolderContext = plot.canvasRenderer.updateCanvasContext(
        geoMap,
        'label-container',
        geoMap.fixIE ? '.static' : 'div.legendarea',
        geoMap.chartParentElement
    );
};

/**
 *
 * @param {String} scopeName
 */
GeoMap.prototype.getLayerData = function(scopeName) {
    var mapCollection = mapCollections;

    if (defined(mapCollection[scopeName])) {
        this.emptyMapLayer = false;
        return mapCollection[scopeName];
    } else {
        this.emptyMapLayer = true;
    }
};

/**
 * function to draw
 */
GeoMap.prototype.setUpMapLayer = function() {
    var geoMap = this,
        systemConfig = geoMap.systemConf,
        dataProcessor = geoMap.dataObject,
        backgroundLayerData;

    if (systemConfig.map.allAreas) {
        let type = geoMap.geo.defaultLayer.layerGeojson.type;
        if (type !== 'FeatureCollection') {
            backgroundLayerData = geoMap.geo.defaultLayer.layerGeojson;
        } else {
            backgroundLayerData = dataProcessor.getMapFeatures('geojson');
        }
    } else {
        backgroundLayerData = dataProcessor.getUserMapFeatures('geojson');
    }

    geoMap.geo.renderedLayer = backgroundLayerData;
    // setup the default layer with the data
    geoMap.setupLayerSettings(backgroundLayerData);
};

/**
 *
 */
GeoMap.prototype.setupZScale = function() {
    var geoMap = this,
        dataProcessor = geoMap.dataObject;

    if (dataProcessor.hasZaxis) {
        geoMap.zscale = geoMap.scales[0].create();
    }
};

GeoMap.prototype.renderGraticule = function() {
    var geoMap = this,
        graticuleConfig = geoMap.systemConf.map.graticule || {},
        graticuleEnabled = graticuleConfig.show === true,
        extraLayers = geoMap.geo.extraLayers;

    if (graticuleEnabled) {
        if (geoMap.updating && extraLayers.graticule) {
            extraLayers.graticule.update();
        } else {
            var graticule = d3_geoGraticule();
            extraLayers.graticule = geoMap.addLayer(graticule(), graticuleConfig, 'graticule', 'extra');
        }
    } else if (defined(extraLayers.graticule)) {
        extraLayers.graticule.destroy();
        extraLayers.graticule = null;
    }
};

GeoMap.prototype.renderGlobe = function() {
    var geoMap = this,
        mapConfig = geoMap.systemConf.map,
        projection = mapConfig.projection || {},
        enabledGlobe = projection.name === 'geoOrthographic' && !defined(mapConfig.globe.show),
        extraLayers = geoMap.geo.extraLayers;

    // temp globe outer border
    if (enabledGlobe || mapConfig.globe.show) {
        if (geoMap.updating && extraLayers.globe) {
            extraLayers.globe.update();
        } else {
            extraLayers.globe = geoMap.addLayer(
                {
                    type: 'Sphere'
                },
                mapConfig.globe,
                'sphere',
                'extra'
            );
        }
    } else if (defined(extraLayers.globe)) {
        extraLayers.globe.destroy();
        extraLayers.globe = null;
    }
};

/**
 *
 */
GeoMap.prototype.renderMapLayer = function() {
    var geoMap = this,
        systemConfig = geoMap.systemConf,
        dataProcessor = geoMap.dataObject,
        userMapConfig = systemConfig.map,
        remainingAreas = dataProcessor.userUnmatchedMapFeatures.values();

    //if the map does not contain heat map render the map as background layer
    if (!dataProcessor.hasHeatmap) {
        // var backgroundLayer = MapDataProcessor.helpers.arrayToGeojson(dataProcessor.userMapFeatures.values());
        var defaultLayer = geoMap.geo.defaultLayer.layerGeojson,
            backgroundLayerFeatures;

        if (defaultLayer.type !== 'FeatureCollection') {
            backgroundLayerFeatures = defaultLayer;
        } else {
            backgroundLayerFeatures = dataProcessor.userMapFeatures.values();
        }
        if (backgroundLayerFeatures.length || backgroundLayerFeatures.type) {
            geoMap.geo.extraLayers.backgroundLayer = geoMap.addLayer(
                backgroundLayerFeatures,
                userMapConfig,
                'background-layer',
                'extra',
                userMapConfig.useMeshOutline
            );
        }
    }

    // fix all area
    if (geoMap.updating && geoMap.geo.extraLayers.allAreas && !geoMap.flushed) {
        geoMap.geo.extraLayers.allAreas.update(remainingAreas);
    } else if (userMapConfig.allAreas) {
        if (!dataProcessor.userUnmatchedMapFeatures.empty()) {
            geoMap.geo.extraLayers.allAreas = geoMap.addLayer(
                remainingAreas,
                userMapConfig,
                'all-areas-layer',
                'extra',
                userMapConfig.useMeshOutline
            );
        }
    }
};

GeoMap.prototype.renderLabels = function() {
    var geoMap = this;
    geoMap.geo.geoLabels = new GeoLabels(geoMap);
};

/**
 *
 */
GeoMap.prototype.handleEvent = function() {
    var geoMap = this,
        geoConfig = geoMap.geo;

    if ((!geoMap.updating || !geoMap.eventHandler) && geoMap.dataObject.totalDataCount) {
        geoMap.createEventGroups();

        // var eventHandler = (geoMap.eventHandler = events.mouseevents(
        //     geoMap,
        //     geoMap.tooltip.element,
        //     geoMap.systemConf,
        //     geoMap.plotarea,
        //     geoMap.dataset,
        //     null,
        //     null,
        //     null,
        //     null,
        //     geoMap.eventWrap,
        //     geoMap.seriesColor
        // ));
        let trackerGroup = createElement('g', '.trackerGroup', [1], geoMap.eventWrap).attr('pointer-events', 'all'),
            trackerElement = createElement('rect', '.eventhandler', [1], trackerGroup);

        var eventHandler = (geoMap.eventHandler = new ZC_MapEvents(geoMap, trackerGroup));

        geoMap.eventHandler.trackerGroup = trackerGroup;
        geoMap.eventHandler.trackerElement = trackerElement;

        eventHandler.init();

        eventHandler.mapEvents = new zmaps.mapEvents(geoMap);

        tooltip.initialize && tooltip.initialize(geoMap);
        let handlers = {
            mousemove: () => {
                if (this.chartRendered) {
                    var isSingleSrcEvt = this.eventHandler.isSingleSource();
                    var node = this.eventHandler.getSourceNode();
                    var mouse = this.eventHandler.getRelativeMousePosition(node, false);
                    return eventHandler.mapEvents.handleEvent(this.eventHandler.mousePos, 'mousemove', this.eventHandler.targetEvent);
                }
            },
            mouseout: () => {
                if (this.chartRendered) {
                    var isSingleSrcEvt = this.eventHandler.isSingleSource();
                    var node = this.eventHandler.getSourceNode();
                    var mouse = this.eventHandler.getRelativeMousePosition(node, false);
                    return eventHandler.mapEvents.handleEvent(null, 'mouseout', this.eventHandler.targetEvent);
                }
            },
            click: () => {
                if (this.chartRendered) {
                    var isSingleSrcEvt = this.eventHandler.isSingleSource();
                    var node = this.eventHandler.getSourceNode();
                    var mouse = this.eventHandler.getRelativeMousePosition(node, false);
                    return eventHandler.mapEvents.handleEvent(this.eventHandler.mousePos, 'click', this.eventHandler.targetEvent);
                }
            },
            doubleclick: () => {
                if (this.chartRendered) {
                    var isSingleSrcEvt = this.eventHandler.isSingleSource();
                    var node = this.eventHandler.getSourceNode();
                    var mouse = this.eventHandler.getRelativeMousePosition(node, false);
                    return eventHandler.mapEvents.handleEvent(this.eventHandler.mousePos, 'doubleclick', this.eventHandler.targetEvent);
                }
            }
        };
        let clickHandlers = {
                click: handlers.click,
                dblclick: handlers.doubleclick,
                contextmenu: handlers.contextmenu
            },
            moveHandlers = {
                mousemove: handlers.mousemove,
                mouseout: handlers.mouseout
            },
            touchHandlers = {
                tap: handlers.click,
                dbltap: handlers.doubleclick,
                singletouchmove: handlers.mousemove,
                touchend: handlers.mouseout
            },
            options = {
                target: 'plot',
                chartObj: this
            };
        // this.eventHandler.handlers = handlers;
        // attachevents(trackerGroup, moveHandlers, null, options, 'chart');
        // attachevents(trackerGroup, clickHandlers, null, options, 'chartmove');
        // attachevents(trackerGroup, {}, touchHandlers, options);
        // trigger init
        // eventHandler.initializeListener();

        geoConfig.eventLayer = eventHandler.trackerElement.styles({
            'pointer-events': 'visible'
        });
    } else if (geoMap.eventHandler) {
        // geoMap.eventHandler.updateParams(geoMap);
    } else {
        return false;
    }

    // define a cache for event params holder
    geoMap.eventHandler.mapEvents.cache = d3_map();

    // drawBorderBox(geoMap.eventHandler.trackerElement, getPlotBoundingRect(geoMap));
    geoMap.positionEventTrackerElement();

    if (geoMap.behaviouralEvents) {
        geoMap.behaviouralEvents.update();
    } else {
        geoMap.behaviouralEvents = new BehaviouralEvents(geoMap);
    }

    geoMap.handleLegendEvents(); // Legend events
};

/**
 *
 */
GeoMap.prototype.clearMap = function() {
    var geoMap = this;
    //clear the canvas before rendering canvas series

    GeoUtils.clearContext(geoMap.geo.featureLayerExtra);
    GeoUtils.clearContext(geoMap.geo.featureSeriesLayer);
    GeoUtils.clearContext(geoMap.geo.geoLabelLayer);
    GeoUtils.clearContext(geoMap.geo.pointLayer);
    GeoUtils.clearContext(geoMap.geo.pointLabelLayer);

    if (defined(geoMap.eventHandler)) {
        geoMap.eventHandler.mapEvents.clearHighlightedPoints();
    }
};

/**
 * Map data process initializer.
 *
 * @param baseSystemConf
 */
GeoMap.prototype.processUserData = function(baseSystemConf) {
    var geoMap = this,
        series,
        systemConfig = (geoMap.systemConf = mergeJSON(baseSystemConf, geoMap.userdata, true, true)),
        userMapConfig = systemConfig.map,
        // get the layer data since we need to map before rendering
        geoLayerData = geoMap.getLayerData(userMapConfig.scope, userMapConfig.layerName);

        let _chartSpecGetter = Registry.getComponent('dataSplicer');
        if (geoMap.userdata.viewParser || geoMap.userdata._viewParser) {
            let _viewParser = Registry.getComponent('viewParser');
            geoMap.viewFlow = new _viewParser(geoMap.userdata.viewParser || geoMap.userdata, geoMap.userdata);
            _chartSpecGetter = Registry.getComponent('dataSplicer4ViewParser');
            geoMap.parserConfig = geoMap.viewFlow.init();
            geoMap.viewFlow.process(geoMap.seriesdata);
        }
    
        geoMap.dataSplicer = new _chartSpecGetter(geoMap.userdata, geoMap.systemConf, geoMap);

    // the first thing to do is call the state change after getting the system config
    if (!geoMap.updating) {
        stateChangeListener(geoMap, geoMap.systemConf.canvas.events, 'chart', 'init');
    }

    series = getPropVal(systemConfig, 'seriesdata') || {};
    geoMap.seriesdata = series.chartdata || [];

    // set the default layer
    var defaultLayer = (geoMap.geo.defaultLayer = {
        layerGeojson: GeoMap.getGeoJSON(geoLayerData, userMapConfig.layerName),
        layerOriginal: geoLayerData
    });

    if (defaultLayer.layerGeojson) {
        geoMap.hasGeoLayer = true;
        geoMap.dataset = new MapDataSet(geoMap);
        geoMap.dataObject = new MapDataProcessor(geoMap);
        geoMap.hasSeriesPoints = geoMap.dataObject.totalDataCount > 0;
        geoMap.nochartdata = !geoMap.hasSeriesPoints && geoMap.emptyMapLayer;
    } else {
        geoMap.hasGeoLayer = false;
        geoMap.nochartdata = true;
    }
};

GeoMap.prototype.addRemoveSeries = function() {
    var geoMap = this,
        eventHandler = geoMap.eventHandler;

    // set updating true for in root level so that every function knows
    geoMap.updating = true;

    geoMap.clearMap();

    // re initialize the procesor;
    geoMap.dataObject.initialize();

    geoMap.updatePlotareaProp();
    geoMap.drawPlotareaClippath();

    geoMap.repaintChart();
};

/**
 *
 */
GeoMap.prototype.repaintChart = function() {
    let geoMap = this;

    // repaint the background layers
    geoMap.repaintGeoLayer();

    //update all series
    geoMap.eventHandler.mapEvents.cache.clear();
    geoMap.eventHandler.oldInfo = null;
    geoMap.plot.renderer.draw();
};

/**
 *
 */
GeoMap.prototype.repaintGeoLayer = function() {
    let geoMap = this;
    each(geoMap.geo.extraLayers, layer => layer && layer.update()); //#ZC1828
    geoMap.geo.customLayers.each(layer => layer && layer.update()); // #ZC1828
    geoMap.geo.geoLabels.update(true);
};

/**
 *
 * @param point
 * @param {number|*} scale
 */
GeoMap.prototype.zoomToPoint = function(point, scale) {
    this.__enableZoom();

    let geoMap = this,
        zoomInstance = this.behaviouralEvents.eventManager.behaviours.PLOT_ZOOM.instance,
        position = geoMap.renderer[point.seriesIndex].getZoomPosition(point);

    ZoomToPosition(position, scale, zoomInstance);
};

/**
 * TODO animation should be in plotoptions level
 * A very basic animation
 */
GeoMap.prototype.animate = function() {
    var geoMap = this,
        mapConfig = geoMap.systemConf.map,
        animationConf = mapConfig.animation,
        selection = geoMap.animateSelection;

    if (selection && !geoMap.updating) {
        selection
            .transition()
            .duration(animationConf.duration)
            .ease(d3_ease(animationConf.easingType))
            .on('start', function() {
                selection.style('opacity', 0);
            })
            .on('end', function() {
                geoMap.chartRendered = true;
            })
            .attrTween('tween', function() {
                return function(t) {
                    selection.style('opacity', t);
                    return t;
                };
            });
    } else {
        geoMap.chartRendered = true;
    }
};

maps.GeoMap = GeoMap;
