// @zohocharts/zc/src/geo-core/map-labels.js
//map-labels.js
import { getFontFamilyByHierarchy, getFontSizeByHierarchy } from '../utils/zcdatautils';
import { pick, invokeFunction, defined, getContrastColor, each, utils } from '../utils/zcutils';
import { isWithinColorRange, getRawColor } from '../utils/colorutils';
import { createElement } from '../utils/zcsvgutils';
import { GeoUtils } from './map-utils';
import { dataLabel_handler } from '../core/main-zc-init';
import { DATALABELS_LABELTYPE } from '../label/labeltype-constants';
import { LatLong } from './coordinates/index';
import { getLabelFontColor } from '../label/datalabelsUtils';
import { isLabelWithinBound, isDatalabelEnabled } from '../label/datalabelsUtils';
import { dataLabelTextContent } from '../label/datalabel-content';
import { _canvasDatalabelUtils } from '../label/canvas-datalabel-utils';
import { d3_select } from '../core/d3-constants';

/**
 *
 * @param {String} text
 * @param {LatLong} latLong
 * @param {Object} styles
 * @param {GeoMap} geoMap
 * @constructor
 */
function Label(text, latLong, styles, geoMap) {
    this.text = text;
    this.latLong = latLong;
    this.styles = styles;
    this.geoMap = geoMap;
}

/**
 *
 * @return {{x: *, y: *}}
 */
Label.prototype.getXY = function() {
    var label = this,
        coordinates = label._coordinates;

    // cache
    if (!coordinates) {
        coordinates = label._coordinates = label.latLong.toXY(label.geoMap);
    }

    return coordinates;
};

/**
 *
 * @return {{width: (*|number), height: *}}
 */
Label.prototype.getDimension = function() {
    var label = this,
        style = label.styles,
        dimension = label._dimension;

    // cache
    if (!dimension) {
        dimension = label._dimension = {
            width: utils.fontUtils.getTextWidth(label.text, style.fontSize, style.fontFamily, style.fontStyle, style.fontWeight),
            height: utils.fontUtils.getTextHeight(label.text, style.fontSize, style.fontFamily, style.fontStyle, style.fontWeight)
        };
    }

    return dimension;
};

/**
 *
 * @return {{x: number, y: number, width: (*|number), height: *}|*}
 */
Label.prototype.getBBox = function() {
    var label = this,
        bbox_geo = label._bbox,
        padding = 5,
        paddingEqual = padding / 2;

    if (!bbox_geo) {
        var xy = label.getXY(),
            dimension = label.getDimension();

        bbox_geo = label._bbox = {
            x: xy.x - dimension.width / 2 - paddingEqual,
            y: xy.y - dimension.height / 2 - paddingEqual,
            width: dimension.width + paddingEqual,
            height: dimension.height + paddingEqual
        };
    }

    return bbox_geo;
};

/**
 *
 * @param {Label} anotherLabel
 */
Label.prototype.overlapsWith = function(anotherLabel) {
    var bbox_geo = this.getBBox(),
        otherBBox = anotherLabel.getBBox();

    return (
        bbox_geo.x < otherBBox.x + otherBBox.width &&
        bbox_geo.x + bbox_geo.width > otherBBox.x &&
        bbox_geo.y < otherBBox.y + otherBBox.height &&
        bbox_geo.height + bbox_geo.y > otherBBox.y
    );
};

/**
 *
 * @param rect
 * @return {boolean}
 */
Label.prototype.insideRect = function(rect) {
    var R2 = this.getBBox();
    return R2.x + R2.width < rect.x + rect.width && R2.x > rect.x && R2.y > rect.y && R2.y + R2.height < rect.y + rect.height;
};

Label.prototype.clearCache = function() {
    this._bbox = this._coordinates = this._dimension = null;
};

/**
 *
 * @param {GeoMap} geoMap
 * @constructor
 */
export function GeoLabels(geoMap) {
    this.geoMap = geoMap;
    this.initialize();
}

GeoLabels.prototype.initialize = function() {
    var geoLabels = this,
        geoMap = geoLabels.geoMap,
        userLabelStyles = geoMap.systemConf.map.labels || {},
        geoLabelStyles = (geoLabels.geoLabelStyles = Object.create(userLabelStyles)),
        labelArray = (geoLabels.labels = []),
        labelBoundary = {
            x: 0,
            y: 0,
            width: geoMap.plotarea.width,
            height: geoMap.plotarea.height
        };

    /**
     *
     * @param {Label} label
     */

    var addLabelToList = function(label) {
        if (
            // ZC1432
            label.insideRect(labelBoundary) &&
            !labelArray.some(function(visibleLabel) {
                return label.overlapsWith(visibleLabel);
            })
        ) {
            labelArray.push(label);
        }
    };

    // series labels
    // loop through series and collect labels
    // each(geoMap.renderer, function(chartSeries, seriesIndex) {
    //
    //     var plopOptions = chartSeries.plotOptions,
    //         dataLabels = plopOptions.datalabels,
    //         chartValue = geoMap.chartTypes[seriesIndex].value;
    //
    //     if (!chartSeries.series.disabled && geoMap.datalabels.moduleExists && isDatalabelEnabled(dataLabels, geoMap) && chartValue === $ZC.charttype.geoheatmap) {
    //
    //         /**
    //          *
    //          * @type {function} getLabelText
    //          */
    //         var getLabelText = dataLabelTextContent(geoMap, null, dataLabels, DATALABELS_LABELTYPE, null, seriesIndex),
    //             seriesLabelStyles = {};
    //
    //         seriesLabelStyles.fontFamily = getFontFamilyByHierarchy(dataLabels, geoMap);
    //         seriesLabelStyles.fontSize = getFontSizeByHierarchy(dataLabels, geoMap);
    //         seriesLabelStyles.fontWeight = dataLabels.fontWeight || "normal";
    //         seriesLabelStyles.fontStyle = dataLabels.fontStyle || "normal";
    //         seriesLabelStyles.textShadow = dataLabels.textShadow;
    //         seriesLabelStyles.strokeColor = pick(dataLabels.strokeColor, userLabelStyles.strokeColor);
    //         seriesLabelStyles.strokeWidth = pick(dataLabels.strokeWidth, userLabelStyles.strokeWidth);
    //         seriesLabelStyles.type = "point";
    //
    //         each(chartSeries.series.data, function(subSeriesPoints, subSeriesIndex) {
    //
    //             each(subSeriesPoints, function(point, itemIndex) {
    //
    //                 var feature = geoMap.dataObject.getMapFeaturesByDataPoint(point, "array")[0];
    //
    //                 if (feature) {
    //
    //                     var text = getLabelText(point),
    //                         label = new Label(text, new LatLong(feature.properties.longitude, feature.properties.latitude), seriesLabelStyles, geoMap),
    //                         rawColor = getRawColor(geoMap, seriesIndex, point),
    //                         render = true;
    //
    //                     // set the font color separately
    //                     label.fontColor = getLabelFontColor(geoMap.systemConf, rawColor, dataLabels, geoMap, false, seriesIndex, false);
    //
    //                     if (geoMap.updating && geoMap.hasColorScale) {
    //                         render = isWithinColorRange(geoMap.dataset.getClr(point), geoMap);
    //                     }
    //
    //                     if (render) {
    //                         addLabelToList(label);
    //                     }
    //                 }
    //
    //             })
    //         })
    //     }
    // });

    // add the default area labels if enabled

    if (geoLabelStyles.show === true) {
        // set geo labels styles
        geoLabelStyles.fontFamily = getFontFamilyByHierarchy(userLabelStyles, geoMap);
        geoLabelStyles.fontSize = getFontSizeByHierarchy(userLabelStyles, geoMap);
        geoLabelStyles.fontWeight = userLabelStyles.fontWeight || 'normal';
        geoLabelStyles.fontStyle = userLabelStyles.fontStyle || 'normal';
        geoLabelStyles.fontColor = userLabelStyles.fontColor || '#333';
        geoLabelStyles.textShadow = userLabelStyles.textShadow;
        geoLabelStyles.strokeColor = userLabelStyles.strokeColor;
        geoLabelStyles.strokeWidth = userLabelStyles.strokeWidth;
        geoLabelStyles.type = 'background';

        // first get series labels if enabled
        if (geoMap.systemConf.map.allAreas === true) {
            geoMap.dataObject.userUnmatchedMapFeatures.each(function(feature, index) {
                var properties = feature.properties,
                    userValue = invokeFunction(true, userLabelStyles.showAs, feature, geoMap),
                    // get the text
                    text = pick(properties[userValue], userValue, properties.name),
                    // create a label instance
                    label = new Label(text, new LatLong(properties.longitude, properties.latitude), geoLabelStyles, geoMap);

                // add
                addLabelToList(label);
            });
        }
    }

    // finally render
    geoLabels.render();
};

/**
 *
 * @param update
 */
GeoLabels.prototype.render = function(update) {
    var geoLabels = this,
        geoMap = geoLabels.geoMap;

    if (this.geoMap.geo.isCanvasMode) {
        geoLabels.renderCanvasLabel(update);
    } else {
        geoLabels.renderSvgLabel(update);
    }
};

GeoLabels.prototype.renderCanvasLabel = function() {
    var geoLabel = this,
        labelLayer = geoLabel.geoMap.geo.geoLabelLayer, // add the labels with the feature itself
        getCanvasFont = function(fontStyle, fontWeight, fontSize, fontFamily) {
            return `${fontStyle} ${fontWeight} ${fontSize}px ${fontFamily}`;
        },
        shadowCache = {};

    geoLabel.labels.forEach(function(label) {
        var styles = label.styles,
            fillColor = pick(label.fontColor, styles.fontColor),
            textStroke = styles.strokeColor || getContrastColor(fillColor),
            coordinates = label.getXY();

        if (defined(coordinates.x) && defined(coordinates.y)) {
            labelLayer.save();

            if (styles.textShadow && !defined(shadowCache[label.type])) {
                shadowCache[label.type] = _canvasDatalabelUtils.parseTextShadow(styles.textShadow);
            }

            if (shadowCache[label.type]) {
                var textShadow = shadowCache[label.type];

                if (defined(textShadow)) {
                    labelLayer.shadowOffsetX = textShadow[0];
                    labelLayer.shadowOffsetY = textShadow[1];
                    labelLayer.shadowBlur = textShadow[2];
                    labelLayer.shadowColor = textShadow[3];
                }
            }

            labelLayer.textAlign = 'center';
            labelLayer.textBaseline = 'middle';
            labelLayer.fillStyle = fillColor;
            labelLayer.strokeStyle = textStroke;
            labelLayer.font = getCanvasFont(styles.fontStyle, styles.fontWeight, styles.fontSize, styles.fontFamily);

            if (styles.strokeWidth > 0) {
                labelLayer.lineWidth = styles.strokeWidth;
                labelLayer.strokeText(label.text, coordinates.x, coordinates.y);
            }

            labelLayer.fillText(label.text, coordinates.x, coordinates.y);
            labelLayer.restore();
        }
    });
};

GeoLabels.prototype.renderSvgLabel = function() {
    var geoLabel = this,
        labelLayer = geoLabel.geoMap.geo.geoLabelLayer;

    // set the style to the parent element
    geoLabel.labelElements = createElement('text', '.geo-labels', geoLabel.labels, labelLayer).each(function(label) {
        var labelElement = d3_select(this),
            coordinates = label.getXY(),
            styles = label.styles,
            fillColor = pick(label.fontColor, styles.fontColor),
            textStroke = styles.strokeColor || getContrastColor(fillColor);

        if (defined(coordinates.x) && defined(coordinates.y)) {
            labelElement.attrs({
                'font-family': styles.fontFamily,
                'font-size': styles.fontSize,
                'font-style': styles.fontStyle,
                'font-weight': styles.fontWeight,
                'text-anchor': 'middle',
                'dominant-baseline': 'middle',
                transform: `translate(${coordinates.x},${coordinates.y})`
            });

            if (styles.textShadow) {
                labelElement.style('text-shadow', styles.textShadow);
            }

            if (styles.strokeWidth > 0) {
                //stroke text
                createElement('tspan', '.stroke-text', [1], labelElement)
                    .text(label.text)
                    .attrs({
                        stroke: textStroke,
                        'stroke-width': styles.strokeWidth,
                        x: 0,
                        y: 0
                    });
            }

            createElement('tspan', '.fill-text', [1], labelElement)
                .text(label.text)
                .attrs({
                    fill: fillColor,
                    x: 0,
                    y: 0
                });
        }
    });
};

GeoLabels.prototype.update = function() {
    var mapLabel = this,
        geoMap = mapLabel.geoMap;

    // fix canvas zoom transforms
    // if (geoMap.geo.isCanvasMode) {
    //     // GeoUtils.clearContext(geoMap.geo.labelLayer);
    // }

    mapLabel.labels.forEach(function(label) {
        label.clearCache();
    });
    mapLabel.initialize(true);
};
