// @zohocharts/zc/src/geo-core/map-utils.js
//GeoUtils.js

/**
 * Group of geo utils.
 *
 * @type {{addLayer: (function(d3.selection, String, Array, d3.selection): *), toCapitalize: (function(*): string), fixHDPI: (function(CanvasRenderingContext2D): CanvasRenderingContext2D)}}
 */
import { d3_geoDistance } from '../core/d3-constants';

export var GeoUtils = {
    /**
     *
     * @param feature
     * @param geoPath
     * @return {*|{x: number, y: number, width: number, height: number}}
     */
    getGeoBBox: function(feature, geoPath) {
        return this.geoBoundToRect(geoPath.bounds(feature));
    },

    /**
     *
     * @param {number[]}bound
     * @return {{x: number, y: number, width: number, height: number}}
     */
    geoBoundToRect: function(bound) {
        return {
            x: bound[0][0],
            y: bound[0][1],
            width: Math.abs(bound[1][0] - bound[0][0]),
            height: Math.abs(bound[1][1] - bound[0][1])
        };
    },

    /**
     *
     * @param feature
     * @param type
     * @param geoPath
     * @return {*|*|{x: number, y: number, width: number, height: number}}
     */
    getFeatureGradientBBox: function(feature, type, geoPath) {
        return this.getGeoBBox(feature, geoPath);
        // var bbox = this.getGeoBBox(feature, geoPath);
        // if (type === "radial") {
        //     bbox.x = bbox.x + bbox.width / 2;
        //     bbox.y = bbox.y + bbox.height / 2;
        // }
        // return bbox;
    },

    /**
     *
     * @param {GeoMap} geoMap
     * @param {number} x
     * @param {number} y
     * @return {number[]}
     */
    toLatLong: function(geoMap, x, y) {
        var utils_geo = geoMap.geo.utils,
            xy = [x, y];

        // first get if any transformation is applied to the map and invert it
        // then invert the xy on transformed container
        // var transform = geoMap.getTransform();
        // xy = transform.invert(xy);

        // finally convert and return
        return utils_geo.projection.invert(xy);
    },

    /**
     *
     * @param {GeoMap}geoMap
     * @param {number} x
     * @param {number} y
     * @return {number[]}
     */
    toXY: function(geoMap, x, y) {
        // fix the globe
        // https://stackoverflow.com/questions/48161441/
        var projectionConf = geoMap.systemConf.map.projection || {};

        if (projectionConf.name === 'geoOrthographic') {
            var rotate = geoMap.geo.utils.projection.rotate(),
                centerPos = [-rotate[0], -rotate[1]];
            var distance = d3_geoDistance([x, y], centerPos);
            if (distance > Math.PI / 2) {
                return [null, null];
            }
        }
        return geoMap.geo.utils.projection([x, y]) || [null, null];
    },

    clearContext: function(context) {
        if (context && context.clearRect) {
            context.clearRect(0, 0, context.canvas.width, context.canvas.height);
        }
    },

    /**
     *
     */
    applyTransform: function(transform, context) {
        if (context.canvas) {
            context.translate(transform.x, transform.y);
            context.scale(transform.k, transform.k);
        } else {
            context.attr('transform', transform);
        }
    },

    round: function(value, precision) {
        var multiplier = Math.pow(10, precision || 0);
        return Math.round(value * multiplier) / multiplier;
    }
};
