// @zohocharts/zc/src/geo-core/map-data-processor.js
//MapDataProcessor.js
import { colorscale_helpers } from '../axis/colorscale-helpers';
import { chartcategory } from '../core/main-zc-init';
import { GeoUtils } from './map-utils';
import { isArray, defined, each, numberIsBetween, extend, isUndefined } from '../utils/zcutils';
import { d3_map, d3_extent, d3_merge, topojson, d3_set, d3_geoCentroid } from '../core/d3-constants';
import { GeoPoint, LatLong } from './coordinates/index';
import { charttype } from '../core/main-zc-init';
import { _bubblePieProcessor } from '../core/data-processor/bubblepie-processor';
import { isBubbleChartType } from '../events/events-utils';
import { encodeMethods } from '../core/data-processor/encode-helpers';
import { getSizeFromPoint } from '../core/data-processor/size/size-from-point';
import { DataProcessor } from '../core/data-processor';
/**
 *
 * @param {GeoMap} geoMap
 * @constructor
 */
export function MapDataProcessor(geoMap) {
    this.geoMap = geoMap;
    this.initialize();
}

MapDataProcessor.prototype.initialize = function() {
    var dataProcessor = this,
        geoMap = dataProcessor.geoMap,
        geoConfig = geoMap.geo,
        dataSet = geoMap.dataset,
        allSeries = geoMap.seriesdata,
        userMapConfig = geoMap.systemConf.map,
        excludeAreas = d3_set(userMapConfig.excludeAreas || []),
        includeAreas = d3_set(userMapConfig.includeAreas || []),
        // get the map features for data mapping
        mapFeatures = (dataProcessor.mapFeatures = geoConfig.defaultLayer.layerGeojson.features),
        joinBy = (dataProcessor.joinByKey = dataSet.getMapByKey()),
        hasLatLongJoinBy = (dataProcessor.hasLatLongJoinBy = dataSet.latLongJoinByKeys.has(joinBy)),
        hashBy = hasLatLongJoinBy ? 'zc-key' : joinBy,
        // hash the map features by joinBy key for faster lookup
        mapFeaturesHashed = (dataProcessor.mapFeaturesHashed = MapDataProcessor.helpers.hashFeaturesByProperty(
            mapFeatures,
            hashBy,
            includeAreas,
            excludeAreas,
            true
        )),
        // hash the data for faster reverse lookup
        userDataHashed = (dataProcessor.userDataHashed = d3_map()),
        // testing
        userUnmatchedMapFeatures = (dataProcessor.userUnmatchedMapFeatures = d3_map()),
        // create hash for faster lookup
        userMapFeatures = (dataProcessor.userMapFeatures = d3_map()),
        // also create a map for unmatched data
        userUnmatchedData = (dataProcessor.userUnmatchedData = d3_set()),
        flattenPoints;

    // total data count;
    dataProcessor.totalDataCount = 0;
    dataProcessor.extents = d3_map();
    dataProcessor.hasPointChart = false;
    dataProcessor.isCombinationalChart = d3_set();
    dataProcessor.flattenPoints = flattenPoints = d3_map();
    dataProcessor.hasHeatmap = dataProcessor.hasBubblePie = dataProcessor.hasPointChart = false;
    dataProcessor.encodes = {};
    dataProcessor.markerSize = {};
    dataProcessor.processedseriesData = allSeries;
    dataProcessor.annotation = [];
    geoMap.cache.data = {};
    geoMap.cache.search = {};

    // set identification
    geoMap.isGeoMap = true;
    geoMap.hasColorScale = colorscale_helpers.isQuantileColorScale(geoMap);

    dataProcessor.hasZaxis =
        defined(dataSet.zidx) &&
        chartcategory.zaxis.some(function(type) {
            return dataProcessor.containsChartType(type);
        });

    // hashing function
    var hash = function(point, seriesIndex, subSeriesIndex, itemIndex) {
        var userJoinBy = dataSet.getGeo(point),
            chartName = geoMap.chartTypes[seriesIndex].name,
            flattenArray = flattenPoints.get(chartName);

        // set index on point
        //Flatten data - for axis chart it was handled in flatten-data.js
        point.seriesIndex = seriesIndex;
        point.subSeriesIndex = subSeriesIndex;
        point.itemIndex = itemIndex;
        point.rendererIndex = seriesIndex;
        point._key = DataProcessor.helpers.generatePointKey(geoMap.id, '0_0', seriesIndex, subSeriesIndex, itemIndex);
        point._data = dataHash(point, seriesIndex, geoMap);

        // flatten array
        if (!defined(flattenArray)) {
            flattenArray = [];
            flattenPoints.set(chartName, flattenArray);
        }
        flattenArray.push(point);

        /**
         * We support geo heatmap with only key value. Not by Lat-Long
         */
        if (!hasLatLongJoinBy || dataProcessor.isValidPoint(point)) {
            if (mapFeaturesHashed.has(userJoinBy)) {
                userMapFeatures.set(userJoinBy, mapFeaturesHashed.get(userJoinBy));
                // hash data
                if (userDataHashed.has(userJoinBy)) {
                    userDataHashed.get(userJoinBy).push(point);
                } else {
                    userDataHashed.set(userJoinBy, [point]);
                }
                // #ZC1652
                // calculate the total data count
                dataProcessor.totalDataCount++;
                saveAnnotation(point, geoMap, dataProcessor.annotation);
            } else if (hasLatLongJoinBy) {
                // calculate the total data count
                // #ZC1652
                dataProcessor.totalDataCount++;
                saveAnnotation(point, geoMap, dataProcessor.annotation);
            } else {
                userUnmatchedData.add(userJoinBy);
            }
        }
    };

    let annotationHash = function(point, seriesIndex, subSeriesIndex, itemIndex) {
        var userJoinBy = dataSet.getGeo(point);

        if (!hasLatLongJoinBy || dataProcessor.isValidPoint(point)) {
            if (mapFeaturesHashed.has(userJoinBy) || hasLatLongJoinBy) {
                saveAnnotation(point, geoMap, dataProcessor.annotation, seriesIndex, subSeriesIndex, itemIndex);
            }
        }
    };

    var dataHash = function(point, seriesIndex, geoMap) {
        return function() {
            return geoMap.renderer[seriesIndex].highlighter.data(point);
        };
    };

    var zExtremes = [];

    // loop through all series
    each(allSeries, function(series, i) {
        // convert to 3D array if needed
        MapDataProcessor.helpers.convertTo3DArray(series);

        var chartType = geoMap.chartTypes[i];

        dataProcessor.isCombinationalChart.add(chartType.name);

        if (!series.disabled) {
            if (!dataProcessor.hasHeatmap && charttype.geoheatmap === chartType.value) {
                dataProcessor.hasHeatmap = dataProcessor.containsChartType('geoheatmap');
            }

            if (!dataProcessor.hasBubblePie && charttype.geobubblepie === chartType.value) {
                dataProcessor.hasBubblePie = true;
            }

            dataProcessor.isBubbleEnabled = dataProcessor.isBubbleEnabled || isBubbleChartType(chartType.value);

            if (!dataProcessor.hasPointChart && charttype.geoheatmap !== chartType.value) {
                dataProcessor.hasPointChart = true;
            }

            each(series.data, function(subSeriesPoints, j) {
                // hash each point
                each(subSeriesPoints, function(point, k) {
                    hash(point, i, j, k);
                });

                // get the extent for bubble chart
                if (dataProcessor.hasZaxis) {
                    dataProcessor.saveEncode('z', 0, i);
                    zExtremes.push(
                        d3_extent(subSeriesPoints, function(point) {
                            return dataSet.getZ(point);
                        })
                    );
                }
            });
        }

        each(series.data, function(subSeriesPoints, j) {
            each(subSeriesPoints, function(point, k) {
                annotationHash(point, i, j, k);
            });
        });
    });

    // filter out the invalid points
    flattenPoints.each(function(points, chartType) {
        flattenPoints.set(
            chartType,
            points.filter(function(point) {
                if (dataProcessor.isValidPoint(point)) {
                    return point;
                }
            })
        );
    });

    // set the final extent
    if (dataProcessor.hasZaxis) {
        dataProcessor.extents.set('z', d3_extent(d3_merge(zExtremes)));
    }

    // for lat-long data type, declare all areas as background layer
    if (hasLatLongJoinBy) {
        dataProcessor.userUnmatchedMapFeatures = mapFeaturesHashed;
    } else {
        mapFeaturesHashed.keys().forEach(function(key) {
            if (!userMapFeatures.has(key) && !excludeAreas.has(key)) {
                userUnmatchedMapFeatures.set(key, mapFeaturesHashed.get(key));
            }
        });
    }

    // process the bubble pie data
    if (dataProcessor.hasBubblePie) {
        // dependency
        dataProcessor.data = geoMap.systemConf;
        dataProcessor.chartObj = geoMap;
        dataProcessor.dataset = dataSet;
        dataProcessor.processedseriesData = geoMap.seriesdata;

        dataProcessor.bubblepieData = dataProcessor.processBubblePieData(null, { chartType: charttype.geobubblepie });
    }

    //finally userMapFeatures save as array for later use
    dataProcessor.isCombinationalChart = dataProcessor.isCombinationalChart.values().length > 1;
    dataProcessor.userMapFeaturesArray = userMapFeatures.values();
    dataProcessor.totalnoofdatas = dataProcessor.totalDataCount;
};

MapDataProcessor.prototype.getFeatureMesh = function(features) {
    var dataProcessor = this,
        tempSet = d3_set(),
        baseTopojsonLayer = dataProcessor.geoMap.geo.defaultLayer.layerOriginal,
        mesh = null,
        layerName = dataProcessor.geoMap.systemConf.map.layerName,
        joinByKey = dataProcessor.hasLatLongJoinBy ? 'zc-key' : dataProcessor.joinByKey;

    if (baseTopojsonLayer.type === 'Topology') {
        features.forEach(function(feature) {
            if (feature.properties && feature.properties[joinByKey]) {
                tempSet.add(feature.properties[joinByKey]);
            }
        });
        mesh = topojson.mesh(baseTopojsonLayer, baseTopojsonLayer.objects[layerName], function(a, b) {
            // #ZC2830
            return tempSet.has(a.properties[joinByKey]) || tempSet.has(b.properties[joinByKey]);
        });
    }
    return mesh;
};

/**
 *
 * @return {boolean}
 */
MapDataProcessor.prototype.getAxisRotated = function() {
    //TODO: remove
    return false;
};

/**
 *
 * @param chartName
 * @return {boolean}
 */
MapDataProcessor.prototype.containsChartType = function(chartName) {
    return this.geoMap.chartTypes.some(function(chartType) {
        return chartType.name === chartName || chartType.value === chartName;
    });
};

/**
 *
 * @param point
 * @return {*}
 */
MapDataProcessor.prototype.getXYPoint = function(point) {
    return this.getLatLonPoint(point).toXY(this.geoMap);
};

/**
 *
 * @param point
 * @return {null|*}
 */
MapDataProcessor.prototype.getFeatureByPoint = function(point) {
    var dataProcessor = this,
        userMapFeatures = dataProcessor.userMapFeatures,
        dataset = dataProcessor.geoMap.dataset,
        key = dataset.getGeo(point);

    if (userMapFeatures.has(key)) {
        return userMapFeatures.get(key);
    }
    return null;
};

/**
 *
 * @param point
 * @return {LatLong}
 */
MapDataProcessor.prototype.getLatLonPoint = function(point) {
    var dataProcessor = this,
        geoMap = dataProcessor.geoMap,
        dataSet = geoMap.dataset,
        userMapFeatures = dataProcessor.userMapFeatures,
        // get the key from the data
        key = dataSet.getGeo(point),
        lonLat = [0, 0];

    if (userMapFeatures.has(key)) {
        var feature = userMapFeatures.get(key),
            properties = feature.properties;

        if (defined(properties.latitude) && defined(properties.longitude)) {
            lonLat = [+properties.longitude, +properties.latitude];
        } else {
            lonLat = d3_geoCentroid(feature);
        }
    } else if (dataProcessor.hasLatLongJoinBy) {
        lonLat = [dataSet.getLongitude(key), dataSet.getLatitude(key)];
    }
    return new LatLong(lonLat[0], lonLat[1]);
};

/**
 *
 * @param axisRotated
 * @param stacked
 * @param xScale
 * @param yScale
 * @param yAxisColumnOrder
 * @param point
 * @param type
 * @return {number}
 */
MapDataProcessor.prototype.getXYvalue = function(axisRotated, stacked, xScale, yScale, yAxisColumnOrder, point, type) {
    var dataProcessor = this;
    if (type === 'z') {
        return dataProcessor.getZFromPoint(point);
    }
    return dataProcessor.getXYPoint(point)[type];
};

/**
 *
 * @param point
 * @return {number}
 */
MapDataProcessor.prototype.getZFromPoint = function(point) {
    let geoMap = this.geoMap;
    return getSizeFromPoint(point, geoMap.dataset, geoMap);
};

/**
 *
 * @param latLong
 * @return {GeoPoint}
 */
MapDataProcessor.prototype.getXYFromLatLng = function(latLong) {
    var xy = GeoUtils.toXY(this.geoMap, latLong.longitude, latLong.latitude);
    return new GeoPoint(xy[0], xy[1]);
};

/**
 *
 * @param type
 * @return {Array|d3.map|{}}
 */
MapDataProcessor.prototype.getUserMapFeatures = function(type) {
    var dataProcessor = this;
    if (type === 'geojson') {
        return MapDataProcessor.helpers.arrayToGeojson(dataProcessor.userMapFeaturesArray);
    } else if (type === 'array') {
        return dataProcessor.userMapFeaturesArray;
    }
    return dataProcessor.userMapFeatures;
};

/**
 *
 * @param dataPoints
 * @param type
 */
MapDataProcessor.prototype.getMapFeaturesByDataPoint = function(dataPoints, type) {
    var dataProcessor = this,
        dataSet = dataProcessor.geoMap.dataset,
        map = d3_map();

    if (!(dataPoints[0] instanceof Array)) {
        dataPoints = [dataPoints];
    }

    dataPoints.forEach(function(point) {
        // dataProcessor.userMapFeatures
        var key = point[dataSet.geoidx];

        if (dataProcessor.userMapFeatures.has(key)) {
            map.set(key, dataProcessor.userMapFeatures.get(key));
        }
    });

    if (type === 'geojson') {
        return MapDataProcessor.helpers.arrayToGeojson(map.values());
    } else if (type === 'array') {
        return map.values();
    }
    return map;
};

/**
 *
 * @param type
 * @return {Array|d3.map|{}}
 */
MapDataProcessor.prototype.getMapFeatures = function(type) {
    var dataProcessor = this;

    var features = d3_map(dataProcessor.userMapFeatures);
    dataProcessor.userUnmatchedMapFeatures.keys().forEach(function(key) {
        features.set(key, dataProcessor.userUnmatchedMapFeatures.get(key));
    });

    if (type === 'geojson') {
        return MapDataProcessor.helpers.arrayToGeojson(features.values());
    } else if (type === 'array') {
        return features;
    }

    return features;
};

/**
 *
 * @return {*}
 */
MapDataProcessor.prototype.getMinMaxZ = function() {
    return this.extents.get('z');
};

/**
 *
 * @return {boolean}
 */
MapDataProcessor.prototype.isstacked = function() {
    return false;
};

MapDataProcessor.prototype.containsNegativeData = function(index) {
    return false;
};

MapDataProcessor.prototype.getScaleProxy = function() {
    var processor = this,
        geoMap = processor.geoMap;

    return {
        xScale: function(point) {
            return processor.getXYPoint(point).x;
        },
        yScale: function(point) {
            return processor.getXYPoint(point).y;
        }
    };
};

/**
 *
 * @return {boolean}
 */
MapDataProcessor.prototype.isValidLatLong = function(lat, long) {
    return numberIsBetween(lat, [-90, 90]) && numberIsBetween(long, [-180, 180]);
};

/**
 *
 * @param point
 * @return {boolean}
 */
MapDataProcessor.prototype.isValidPoint = function(point) {
    var processor = this,
        geoIndex = processor.geoMap.dataset.getGeo(point);

    if (!point.length || !defined(geoIndex) || !geoIndex.length) {
        return false;
    }

    if (processor.hasLatLongJoinBy) {
        var lat = processor.geoMap.dataset.getLatitude(geoIndex),
            long = processor.geoMap.dataset.getLongitude(geoIndex);
        return defined(lat) && defined(long) && processor.isValidLatLong(lat, long);
    } else {
        return processor.userDataHashed.has(geoIndex);
    }
};

/**
 *
 * @param {Array} points
 * @return {Array}
 */
MapDataProcessor.prototype.updateSeriesdatabyCategories = function(points) {
    var processor = this,
        ii = 0;
    // ,
    // currentViewPort = processor.geoMap.navigation && processor.geoMap.navigation.getCurrentViewPort();

    return points.filter(function(point) {
        var isValidPoint = processor.isValidPoint(point);

        // TODO remove points that are out of the viewport
        // if (isValidPoint) {
        //     if (currentViewPort) {
        //         let xy = processor.getXYPoint(point);
        //         isValidPoint =  (currentViewPort[0][0] <= xy.x && xy.x <= currentViewPort[1][0]) && (currentViewPort[0][1] <= xy.y && xy.y <= currentViewPort[1][1]);
        //     }
        // }

        if (isValidPoint) {
            point.itemIndex = ii++;
        }
        return isValidPoint;
    });
};

/**
 *
 * @type {DataProcessor.processBubblePieData}
 */
MapDataProcessor.prototype.processBubblePieData = _bubblePieProcessor;

// *************************************************************************** //

MapDataProcessor.helpers = {
    /**
     *
     * @param data
     * @return {*}
     */
    is3D: function(data) {
        return isArray(data) && isArray(data[0]) && isArray(data[0][0]);
    },

    /**
     *
     * @param series
     * @return {Array}
     */
    convertTo3DArray: function(series) {
        if (!this.is3D(series.data)) {
            series.data = [series.data];
        }
    },

    arrayToGeojson: function(features) {
        return {
            type: 'FeatureCollection',
            features: features
        };
    },

    /**
     *
     * @param object
     * @param property
     * @param includeAreas
     * @param excludeAreas
     * @param copy
     * @return {d3.map}
     */
    hashFeaturesByProperty: function(object, property, includeAreas, excludeAreas, copy) {
        var hash = d3_map();
        each(object, function(feature, i) {
            var key = feature.properties[property] || i;
            if ((includeAreas.empty() || includeAreas.has(key)) && (excludeAreas.empty() || !excludeAreas.has(key))) {
                hash.set(key, copy ? Object.create(feature) : feature);
            }
        });
        return hash;
    }
};

function saveAnnotation(curdata, chart, arr, seriesIndex, subSeriesIndex, itemIndex) {
    let config = chart.systemConf.notes || {},
        { dataIndex } = config;

    if (isUndefined(dataIndex)) {
        return;
    }

    let data = curdata;
    if (data[dataIndex]) {
        arr.push({ data: curdata, seriesIndex, subSeriesIndex, itemIndex });
    }
}

extend(MapDataProcessor.prototype, encodeMethods);
