// @zohocharts/zc/src/geo-core/map-renderer.js
import { GeoSeriesTypes } from '../core/main-zc-init';
import { each, arrayIncludes, pick, defined, isUndefined } from '../utils/zcutils';
import { stateChangeListener } from '../events/eventActions';
import { renderer_helpers, Console } from '../core/main-zc-init';
import { d3_scaleLinear } from '../core/d3-constants';
import { plot } from '../core/main-zc-init';
import { chartcategory, charttype } from '../core/main-zc-init';

/**
 *
 * @param {GeoMap} geoMap
 * @constructor
 */
export function GeoPlotRenderer(geoMap) {
    this.geoMap = geoMap;
    this.initialize();
}

GeoPlotRenderer.prototype.initialize = function() {
    var renderer = this,
        geoMap = renderer.geoMap,
        mode = geoMap.systemConf.chart.plot.renderer.mode.toLowerCase();

    // if the mode is canvas check for support
    each(geoMap.chartTypes, function(chartType) {
        if (mode === 'canvas' && arrayIncludes(chartcategory.svgOnly, chartType.value)) {
            mode = 'svg';
            Console.warn(`Fallback: ${chartType.name} was rendered on svg mode`);
        }
    });
    geoMap.renderer = [];
    geoMap.rendererConf = [];

    renderer.mode = mode;
    renderer.MODE = mode.toUpperCase();

    geoMap.extraParams = { labelQueue: [], notesArray: [], labelArray: {}, barDatasetIndex: 0 };
};

/**
 *
 */
GeoPlotRenderer.prototype.draw = function() {
    var plotRenderer = this,
        geoMap = plotRenderer.geoMap,
        systemConfig = geoMap.systemConf;

    geoMap.chartRendered = false;

    // fire plot start event on first load
    // #ZC1732
    stateChangeListener(geoMap, systemConfig.chart.plot.events, 'plot', 'init');

    //setup z scale
    geoMap.setupZScale();

    plotRenderer.renderSeries();

    plotRenderer.transformWrap();
    renderer_helpers.renderCompletedEvents(geoMap, geoMap.updating, geoMap.extraParams, function() {
        if (!geoMap.updating) {
            geoMap.animate();
            geoMap.updating = false;
        } else {
            // ZC1740
            geoMap.chartRendered = true;
        }
    });
};

GeoPlotRenderer.prototype.update = function() {
    this.draw();
};

/**
 * render all the series
 */
GeoPlotRenderer.prototype.renderSeries = function() {
    var plotRenderer = this,
        geoMap = plotRenderer.geoMap,
        systemConfig = geoMap.systemConf,
        allChartSeries = renderer_helpers.arrangeSeriesRenderingOrder(geoMap) || [],
        chartSeries,
        scaleProxy = d3_scaleLinear();

    geoMap.rendererArranged = [];

    each(allChartSeries, function(series, index) {
        var seriesIndex = pick(series.seriesidx, index),
            chartType = geoMap.chartTypes[seriesIndex];

        // define the bubble pie renderer index
        if (isUndefined(geoMap.bubblePieRendererIndex) && chartType.value === charttype.geobubblepie) {
            geoMap.bubblePieRendererIndex = seriesIndex;
        }

        chartSeries = geoMap.renderer[seriesIndex];

        if (!defined(chartSeries) && GeoSeriesTypes.has(chartType.name)) {
            chartSeries = new (GeoSeriesTypes.get(chartType.name))(series, geoMap, seriesIndex);
            geoMap.renderer[seriesIndex] = chartSeries;
            geoMap.rendererConf[seriesIndex] = new plot.rendererConf();
        }

        if (chartType.name === 'geoheatmap') {
            renderer_helpers.setRendererProp(geoMap, chartSeries, series, seriesIndex, 0, geoMap.extraParams);
        }

        geoMap.rendererArranged.push(chartSeries);

        // fix the scale dependency with a proxy.
        scaleProxy.origRange = [0, 0];
        geoMap.rendererConf[seriesIndex]
            .xscale(scaleProxy)
            .yscale(scaleProxy)
            .datasetIndex(seriesIndex)
            .legendOrderIndex(seriesIndex)
            .rendererIndex(seriesIndex);

        clearTimeout(chartSeries.renderedTimer);

        chartSeries.render(geoMap.updating);
    });

    geoMap.rendererArrangedReveresd = geoMap.rendererArranged.slice(0).reverse();
};

/**
 *
 * @param {GeoHeatMapSeries|GeoBubbleSeries|MapSeries} mapSeries
 */
GeoPlotRenderer.prototype.triggerChartRendering = function(mapSeries) {
    var plotRenderer = this,
        geoMap = plotRenderer.geoMap,
        series = mapSeries.series,
        seriesIndex = mapSeries.index,
        chartName = geoMap.chartTypes[seriesIndex].name;

    if (!defined(mapSeries.chartRenderer)) {
        mapSeries.chartRenderer =
            plotRenderer.MODE === 'CANVAS'
                ? plot.canvasRenderer.createCanvasRenderer(chartName)
                : plot.svgRenderer.createSVGRenderer(chartName, geoMap);
    }

    // hide it
    geoMap.container.select('.highlight-canvas').style('visibility', 'hidden');

    mapSeries.commonRendererProp = mapSeries.chartRenderer.commonRendererProp;
    renderer_helpers.setRendererProp(geoMap, mapSeries.chartRenderer, series, seriesIndex, 0, geoMap.extraParams);

    // assign the chartRendered property in the new instance
    if (!mapSeries.hasOwnProperty('chartRendered')) {
        Object.defineProperty(mapSeries, 'chartRendered', {
            get: function() {
                return mapSeries.chartRenderer.chartRendered;
            }
        });
    }

    // finally render it
    plotRenderer[`render${plotRenderer.MODE}`](mapSeries);

    // map the seriesdataAfterUpdate
    if (!mapSeries.hasOwnProperty('seriesdataAfterUpdate')) {
        Object.defineProperty(mapSeries, 'seriesdataAfterUpdate', {
            get: function() {
                return pick(mapSeries.chartRenderer.seriesdataAfterUpdate);
            }
        });
    }
};

/**
 *
 * @param {GeoScatterSeries|GeoBubbleSeries|GeoBubblePieSeries} mapSeries
 */
GeoPlotRenderer.prototype.renderSVG = function(mapSeries) {
    var plotRenderer = this,
        geoMap = plotRenderer.geoMap,
        series = mapSeries.series;
    plot.svgRenderer.drawSVG(geoMap, geoMap.seriesdata, mapSeries.chartRenderer);
};

/**
 *
 * @param {GeoScatterSeries|GeoBubbleSeries} mapSeries
 */
GeoPlotRenderer.prototype.renderCANVAS = function(mapSeries) {
    var plotRenderer = this,
        geoMap = plotRenderer.geoMap,
        series = mapSeries.series;

    plot.canvasRenderer.drawCanvas(geoMap, series, mapSeries.chartRenderer);
};

GeoPlotRenderer.prototype.transformWrap = function() {
    var geoMap = this.geoMap,
        plotArea = geoMap.plotarea;

    // label container
    geoMap.staticPaneSVG.select('.labelholder').attr('transform', `translate(${plotArea.left},${plotArea.top})`);
};
