// @zohocharts/zc/src/geo-core/map-events.js
//map-events.js
import { legend_helpers, zmaps } from '../core/main-zc-init';
import { pick, defined, invokeFunction } from '../utils/zcutils';
import { createElement, appendEle } from '../utils/zcsvgutils';
import { getTooltipPosition } from '../events/events-utils';
// import { legend_helpers } from '../legend/legend-helper';
import { MapLayer } from './map-layer';
import { getEventConf, getEventObject } from '../events/events-utils';
import { $Browser } from '../core/browser-detect';
import { getRawColor } from '../utils/colorutils';
import { GeoPoint, LatLong } from './coordinates/index';
// import {GeoHeatMapSeries} from "../geo-heatmap/geo-heatmap";
import { TRANSPARENT, NONE } from '../core/zc-constants';
import { isBubbleChartType } from '../events/events-utils';
import { d3_namespace, d3_polygonContains, d3_zoomTransform, d3_event, d3_quadtree, d3_select } from '../core/d3-constants';
import { hideCursor_C, showCursor_C } from '../events/visualEffects/controllers/cursor_c';
import { clearHighlightContext } from '../events/visualEffects/highlight/highlight-context';
import { triggerUserCallback } from '../events/customCallbackUtils';
import { assignTouchEventLog } from '../events/listeners/click';
import { getChartTypeByIndex } from '../core/charttype-helpers';

/**
 *
 * @param {GeoMap} geoMap
 * @constructor
 */
export function MapEvents(geoMap) {
    this.geoMap = geoMap;
    this.initialize();
}

/**
 *
 * @param geoMap
 * @param coordinates
 * @return {*}
 */
MapEvents.fixMouseCoordinates = function(geoMap, coordinates) {
    if (!coordinates) {
        return {
            x: null,
            y: null
        };
    }

    return {
        x: coordinates[0],
        y: coordinates[1]
    };
};

/**
 *
 * @param dataPoint
 * @return {{seriesPos: *, itemPos: *, subseriespos: *, data: *}}
 */
MapEvents.fixEventArguments = function(dataPoint) {
    return {
        seriesPos: dataPoint.seriesIndex,
        itemPos: dataPoint.itemIndex,
        subseriespos: dataPoint.subSeriesIndex,
        data: dataPoint
    };
};

MapEvents.prototype.initialize = function() {
    var mapEvents = this,
        geoMap = mapEvents.geoMap,
        mapConfig = geoMap.geo,
        utils_map = (mapEvents.utils = mapConfig.utils),
        userMapConfig = geoMap.systemConf.map;

    mapEvents.eventConfig = pick(geoMap.systemConf.map.events, {});
    mapEvents.outerGeoBounds = utils_map.geoPath.bounds(mapConfig.defaultLayer.layerGeojson);

    var highlightElementParent = createElement('g', '.zcMapClipHighlight', [1], geoMap.staticPaneSVG, ':first-child').attr(
        'clip-path',
        `url(${geoMap.defsLocation}#highlight_group_clip_${geoMap.id}`
    );

    mapEvents.highlightElement = createElement('g', '.map-highlight', [1], highlightElementParent).attr(
        'transform',
        `translate(${geoMap.plotarea.left},${geoMap.plotarea.top})`
    );

    // to save timers
    mapEvents.timers = {};

    mapEvents.highlightCache = {};
    mapEvents.outLine = mapEvents.getOutLine();
};

/**
 *
 * @param position
 * @param point
 * @param event
 * @param pointInfo
 * @param isSamePoint
 */
MapEvents.prototype.showTooltip = function(position, point, event, pointInfo, isSamePoint) {
    var mapEvents = this,
        geoMap = mapEvents.geoMap,
        XY = { x: 0, y: 0 },
        tooltipPosition,
        helpers,
        isOrdinal = true,
        seriesIndex = pointInfo.point.seriesIndex,
        chartType = geoMap.chartTypes[seriesIndex].value,
        plotOptions = geoMap.renderer[seriesIndex].plotOptions;

    if (pointInfo.type === 'feature' || !isSamePoint) {
        if (pointInfo.type === 'point') {
            isOrdinal = false;
            XY = geoMap.dataObject.getXYPoint(point);
        }

        tooltipPosition = getTooltipPosition(
            getEventObject(event),
            position,
            XY.x,
            XY.y,
            geoMap,
            point,
            false,
            false,
            {},
            isOrdinal,
            false,
            false
        );
        helpers = tooltipPosition.helpers;

        // if (chartType === $ZC.charttypes.geobubble || chartType === $ZC.charttypes.geobubble)

        if (isBubbleChartType(chartType)) {
            tooltipPosition.distance = geoMap.dataObject.getXYvalue(false, false, null, null, null, point, 'z') / 2;
        } else if (pointInfo.type === 'point' && true) {
            tooltipPosition.distance = (plotOptions.marker.size - 1) * 10;
        }

        helpers.itempos = point.itemIndex;
        helpers.subseriespos = point.subSeriesIndex;
        helpers.seriesindex = point.seriesIndex;

        geoMap.eventHandler.tooltipElement.show(geoMap, tooltipPosition, null, true, tooltipPosition.helpers);
    }
};

/**
 *
 */
MapEvents.prototype.hideTooltip = function(immediate) {
    var mapEvents = this,
        delay = mapEvents.geoMap.systemConf.tooltip.hideDelay,
        tooltip = mapEvents.geoMap.eventHandler.tooltipElement;

    if (immediate) {
        tooltip.hide(tooltip.element);
    } else {
        clearTimeout(mapEvents.timers.tooltipDelay);
        mapEvents.timers.tooltipDelay = setTimeout(function() {
            tooltip.hide(tooltip.element);
        }, pick(delay, 150));
    }
};

/**
 *
 * @param {GeoPoint|LatLong|[number,number]} latLong
 */
MapEvents.prototype.getPoint = function(latLong) {
    var mapEvents = this,
        geoMap = mapEvents.geoMap,
        point = null,
        xy;

    if (latLong instanceof GeoPoint) {
        xy = latLong.toArray();
        latLong = latLong.toLatLong(mapEvents.geoMap).toArray();
    } else if (latLong instanceof LatLong) {
        xy = latLong.toXY(geoMap);
        latLong = latLong.toArray();
    }

    //check if the mouse is inside the actual map bounds
    var insideMapBound = mapEvents.outLine.some(function(outline) {
        return d3_polygonContains(outline, xy);
    });
    if (insideMapBound) {
        geoMap.rendererArrangedReveresd.some(function(series) {
            // check if event is available for the current series
            if (!geoMap.eventHandler.eventDisabled(series.series) && series.getPoint) {
                return (point = series.getPoint(latLong, xy, series.index));
            }
        });
    }

    return point;
};

MapEvents.prototype.getPointsByIndex = function(seriesIndex, legendFilterIndex) {
    var mapEvents = this,
        geoMap = mapEvents.geoMap,
        points = [];

    if (defined(legendFilterIndex)) {
        var flattenPoints = geoMap.dataObject.flattenPoints.get('geoheatmap');
        points = legend_helpers.filterSelectionByColor(flattenPoints, true, legendFilterIndex, geoMap);
    } else if (defined(seriesIndex)) {
        geoMap.seriesdata[seriesIndex].data.forEach(function(subPoint) {
            points = points.concat(subPoint);
        });
    }

    return points;
};

/**
 *
 * @param seriesIndex
 * @param color
 * @param type
 * @param from
 * @return {{fillColor}}
 */
MapEvents.prototype.getHighlightEffect = function(seriesIndex, color, type, from) {
    var geoMap = this.geoMap,
        highlightEffect = geoMap.legendHighlight.getLegendHighlightEffect(seriesIndex, type, from),
        highlightStyle = {
            fillColor: TRANSPARENT
        };

    if (highlightEffect !== NONE) {
        if (geoMap.fixIE) {
            color = 'transparent';
        }

        highlightStyle = geoMap.legendHighlight.getEffect(
            highlightEffect,
            'area',
            color,
            {
                fill: TRANSPARENT,
                stroke: color
            },
            0,
            from,
            null,
            getChartTypeByIndex(geoMap, '0_0', seriesIndex).name
        );

        // #ZC2361
        highlightStyle = { ...highlightStyle, strokeWidth: pick(highlightStyle.strokeWidth, 2), fillColor: highlightStyle.fill };
    }

    return highlightStyle;
};

/**
 *
 * @param points
 * @param type
 * @param styles
 */
MapEvents.prototype.highlightPoints = function(points, type, styles) {
    var mapEvents = this,
        geoMap = mapEvents.geoMap,
        scatterWrap = geoMap.eventWrap.selectAll('#zcscatterhighlight'),
        featureWrap = geoMap.staticPaneSVG.selectAll('g.map-highlight');
    clearHighlightContext(geoMap);
    switch (type) {
        case 'feature': {
            scatterWrap.attr('visibility', 'hidden');
            featureWrap.attr('visibility', 'visible');
            mapEvents.highlightGeoFeature(points, styles);
            break;
        }
        case 'point': {
            scatterWrap.attr('visibility', 'visible');
            featureWrap.attr('visibility', 'hidden');

            // remove existing point
            scatterWrap.selectAll('*').remove();
            scatterWrap.attr('visibility', 'visible');

            var seriesIndex = points.seriesIndex,
                subSeriesIndex = points.subSeriesIndex,
                itemIndex = points.itemIndex,
                chartType = geoMap.chartTypes[seriesIndex],
                xy = geoMap.dataObject.getXYPoint(points),
                plotOptions = geoMap.seriesdata[seriesIndex].plotoptions,
                commonPlotOptions = geoMap.systemConf.chart.plot.plotoptions[chartType.name],
                currentWrap = appendEle(scatterWrap, 'g', [1], '.', 'class', `highlight${seriesIndex}`),
                renderer = chartType.name == 'geobubblepie' ? geoMap.renderer[geoMap.bubblePieRendererIndex] : geoMap.renderer[seriesIndex],
                highlighter = renderer.chartRenderer.getHighlighter(geoMap),
                defaultH = highlighter._defaultH();
            points._x = xy.x;
            points._y = xy.y;

            points._size = renderer.chartRenderer.getSymbolSize(points);
            defaultH[0].path.call(highlighter, points, defaultH[0]);
            // geoMap.eventHandler.hoverScatter(
            //     currentWrap,
            //     chartType.value,
            //     plotOptions,
            //     commonPlotOptions,
            //     xy.x,
            //     xy.y,
            //     null,
            //     null,
            //     null,
            //     geoMap.seriesdata,
            //     subSeriesIndex,
            //     seriesIndex,
            //     itemIndex,
            //     false,
            //     null,
            //     false
            // );
        }
    }
};

/**
 *
 * @param points
 * @param styles
 */
MapEvents.prototype.highlightGeoFeature = function(points, styles) {
    var mapEvents = this,
        geoMap = mapEvents.geoMap;

    mapEvents.highlightLayer = new MapLayer(points, geoMap.geo.highlightUtils, mapEvents.highlightElement, styles, geoMap);
};

/**
 *
 */
MapEvents.prototype.clearHighlightedPoints = function() {
    this.geoMap.staticPaneSVG.selectAll('g#zcscatterhighlight, g.map-highlight').attr('visibility', 'hidden');
};

/**
 *
 * @return {d3.transform}
 */
MapEvents.prototype.getTransform = function() {
    return d3_zoomTransform(this.geoMap.geo.eventLayer.node());
};

/**
 *
 * @param coordinates
 * @param type
 * @param event
 */
MapEvents.prototype.handleEvent = function(coordinates, type, event) {
    // fix the coordinates
    coordinates = MapEvents.fixMouseCoordinates(this.geoMap, coordinates);
    event = event || d3_event();

    //***************************************************//

    var mapEvents = this,
        geoMap = mapEvents.geoMap,
        geoPoint = new GeoPoint(coordinates.x, coordinates.y),
        point = mapEvents.getPoint(geoPoint),
        tooltipEnabled = geoMap.systemConf.tooltip.enabled !== false, // show by default
        actualEventType = pick(geoMap.zcEventType, type),
        oldPoint = geoMap.eventHandler.oldInfo,
        isSamePoint = !(!oldPoint || oldPoint.point !== (point && point.point));

    if (type === 'mouseout') {
        // get the point for mouse out
        point = geoMap.eventHandler.oldInfo;
    } else if (
        // mouseout if previously it had points and then second time no point is available.
        (defined(geoMap.eventHandler.oldInfo) && !defined(point)) ||
        // mouseout on doubletap or if the same point is being tapped twice.
        ((actualEventType === 'doubletap' || (actualEventType === 'tap' && isSamePoint)) && type !== 'mouseout')
    ) {
        return geoMap.eventHandler.mapEvents.handleEvent(null, 'mouseout', geoMap.eventHandler.targetEvent);
        // return geoMap.eventHandler.mouseOut();
    }

    // if no point return
    if (!defined(point)) {
        return;
    }

    var dataPoint = point.point,
        seriesIndex = dataPoint.seriesIndex,
        mapSeries = geoMap.renderer[seriesIndex],
        plotOptions = mapSeries.plotOptions,
        customUserCallback = getEventConf(null, plotOptions, geoMap, actualEventType),
        eventHelpers = geoMap.eventHandler,
        immediateMouseOut = false,
        actionType = type,
        isTouchDevice = $Browser.isTouchDevice,
        defaultEventHash = {
            mousemove: 'handleMouseMove',
            click: 'handleMouseClick',
            mouseout: 'handleMouseOut'
        };

    // we are showing point on tap for touch devices
    if (isTouchDevice) {
        immediateMouseOut = true;
        // changing the actionType to mousemove for default action
        if (type === 'click') {
            assignTouchEventLog(actualEventType, geoMap);
            actionType = 'mousemove';
        }
    }

    // invoke the user custom function
    if (defined(customUserCallback)) {
        var action = triggerUserCallback(customUserCallback, point, event, null, geoMap);
        invokeFunction(action.callback, geoMap.eventHandler.targetEvent, action.args, geoMap);
    }

    /**
     * Handle default events here for all the event types (mouse & touch)
     */
    if (eventHelpers.eventManager.isDefaultAllowed(customUserCallback, event)) {
        if (defined(mapEvents[defaultEventHash[actionType]])) {
            mapEvents[defaultEventHash[actionType]](point, isSamePoint);
        }
    }

    // handle the mouse move
    if (actionType === 'mousemove') {
        // tooltip show of enabled
        if (tooltipEnabled) {
            clearTimeout(mapEvents.timers.tooltipDelay);
            mapEvents.showTooltip(coordinates, dataPoint, event, point, isSamePoint);
        }

        //add the cursor only once for a particular point
        if (!isSamePoint) {
            showCursor_C.call(
                geoMap.eventHandler.eventManager,
                coordinates,
                { nearest: dataPoint },
                getEventConf(null, mapSeries.plotOptions, geoMap, 'cursor')
            );
            // geoMap.eventHandler.addCursor(getEventConf(null, mapSeries.plotOptions, geoMap, 'cursor'));
            // save the point
            geoMap.eventHandler.eventManager.selectionManager.nearest = dataPoint;
            geoMap.eventHandler.oldInfo = point;

            geoMap.eventHandler.oldInfo.nearestPoint = MapEvents.fixEventArguments(dataPoint);
        }

        geoMap.eventHandler.reverseLegendHighlight.highlight(dataPoint);
        geoMap.eventHandler.unselectedHighlight.highlight(dataPoint);
    }
    // handle the mouse out
    else if (actionType === 'mouseout') {
        // hide show of enabled
        if (tooltipEnabled) {
            mapEvents.hideTooltip(immediateMouseOut);
        }

        // hide the cursor
        // geoMap.eventHandler.hideCursor();
        hideCursor_C.call(geoMap.eventHandler.eventManager);
        //ZC1434
        geoMap.eventHandler.reverseLegendHighlight.reset();
        geoMap.eventHandler.oldInfo = null; // clear it
    }
};

MapEvents.prototype.getOutLine = function() {
    var mapEvents = this,
        geoMap = mapEvents.geoMap,
        outLinePath = document.createElementNS(d3_namespace('svg').space, 'path'),
        tempPath = d3_select(outLinePath)
            .datum({
                type: 'Sphere'
            })
            .attr('d', geoMap.geo.highlightUtils.geoPath),
        outline = tempPath.attr('d');

    return outline
        .split('M')
        .slice(1)
        .map(function(segment) {
            segment = segment.slice(0, segment.length - 1);
            return segment.split('L').map(function(value) {
                return value
                    .trim()
                    .split(/[,\s]/)
                    .map(function(value1) {
                        return parseInt(value1);
                    });
            });
        });
};

/**
 *
 * @param points
 * @return {d3.quadtree}
 */
MapEvents.prototype.getTreeDiagram = function(points) {
    var mapEvents = this,
        geoMap = mapEvents.geoMap;

    return d3_quadtree()
        .x(function(d) {
            return geoMap.dataObject.getXYPoint(d).x;
        })
        .y(function(d) {
            return geoMap.dataObject.getXYPoint(d).y;
        })
        .extent([
            [0, 0],
            [geoMap.plotarea.width, geoMap.plotarea.height]
        ])
        .addAll(points);
};

/**
 *
 * @type {{getHighlightStyle: MapEvents.helpers.getHighlightStyle, handleMouseClick: MapEvents.helpers.handleMouseClick, handleMouseMove: MapEvents.helpers.handleMouseMove, handleMouseOut: MapEvents.helpers.handleMouseOut}}
 */
MapEvents.prototype.handleMouseMove = function(pointInfo, isSamePoint) {
    var mapEvents = this,
        geoMap = mapEvents.geoMap,
        point = pointInfo.point,
        seriesIndex = point.seriesIndex;

    if (!isSamePoint) {
        var fillColor = getRawColor(geoMap, seriesIndex, point),
            highlightStyle = mapEvents.getHighlightEffect(seriesIndex, fillColor, 'selectedSeries', 'plot');

        if (pointInfo.type === 'feature') {
            point = pointInfo.feature;
        }
        mapEvents.highlightPoints(point, pointInfo.type, highlightStyle);
    }
};

MapEvents.prototype.handleMouseOut = function(point) {
    var mapEvents = this,
        geoMap = this.geoMap;

    mapEvents.clearHighlightedPoints();
    hideCursor_C.call(geoMap.eventHandler.eventManager);
};

MapEvents.prototype.handleMouseClick = function(point) {
    this.geoMap.eventHandler.mapEvents.hideTooltip(true);
};

zmaps.mapEvents = MapEvents;
