// @zohocharts/zc/src/geo-core/map-layer.js
//MapLayer.js
import { splat, pick, defined, toCapitalize, each } from '../utils/zcutils';
import { getGradients } from '../utils/colorutils';
import { GeoUtils } from './map-utils';
import { getDashArray } from '../utils/dashstyles';
import { createElement } from '../utils/zcsvgutils';
import { d3_set } from '../core/d3-constants';

/**
 *
 * @param {Object} layer
 * @param {{projection: function, geoPath: d3.geoPath}} util
 * @param {SVGGElement|CanvasRenderingContext2D} layerContainer
 * @param {styles} styles
 * @param {GeoMap} geoMap
 * @param {boolean} useMesh
 * @constructor
 */
export function MapLayer(layer, util, layerContainer, styles, geoMap, useMesh) {
    this.layer = layer;
    this.utils = util;
    this.layerContainer = layerContainer;
    this.styles = styles;
    this.geoMap = geoMap;
    this.useMesh = useMesh;
    this.initialize();
}

/**
 * add supported types
 */
MapLayer.GEO_FORMATS = d3_set([
    'Point',
    'MultiPoint',
    'LineString',
    'MultiLineString',
    'Polygon',
    'MultiPolygon',
    'GeometryCollection',
    'Feature',
    'FeatureCollection',
    'Topology'
]);

/**
 *
 * @return {*}
 */
MapLayer.getMapFeatures = function(object) {
    if (object.type === 'FeatureCollection') {
        return object.features;
    } else {
        return splat(object);
    }
};

/**
 *
 * @param feature
 * @param property
 * @param checkForGradient
 * @returns {*}
 */
MapLayer.prototype.getStyle = function(feature, property, checkForGradient) {
    var mapLayer = this,
        styles = mapLayer.styles || {},
        featureStyle = feature.styles || {},
        properties = feature.properties || {};

    var styleValue = pick(styles[property], featureStyle[property], properties[property], mapLayer.defaultStyles[property]);

    if (typeof styleValue === 'function') {
        return styleValue.call(this, feature, styleValue);
    }

    // check gradient if needed
    if (checkForGradient) {
        if (!mapLayer._gradient) {
            var gradients = pick(styles.gradients, featureStyle.gradients, properties.gradients, {});
            if (gradients.type === 'linear' || gradients.type === 'radial') {
                var bbox_geo = GeoUtils.getFeatureGradientBBox(feature, gradients.type, mapLayer.utils.geoPath);
                styleValue = getGradients(
                    gradients,
                    [bbox_geo.width, bbox_geo.height],
                    mapLayer.getStyle(feature, 'fillColor'),
                    mapLayer.renderingMode,
                    mapLayer.fillOpacity,
                    'area',
                    null,
                    mapLayer.geoMap
                );
                feature.transform = {
                    x: bbox_geo.x,
                    y: bbox_geo.y
                };

                //cache for svg because it uses relative coordinate for each element
                if (!mapLayer.isCanvasMode && gradients.type === 'linear') {
                    mapLayer._gradient = styleValue;
                }
            }
        } else {
            styleValue = mapLayer._gradient;
        }
    } else if (property === 'dashStyle' && defined(styleValue)) {
        styleValue = getDashArray(styleValue, mapLayer.getStyle('strokeWidth'));
    }

    return styleValue;
};

MapLayer.prototype.initialize = function() {
    var mapLayer = this,
        renderingMode = mapLayer.layerContainer.isD3Selection ? 'svg' : 'canvas';

    // set a default styles
    mapLayer.defaultStyles = {
        fillColor: 'transparent',
        strokeColor: '#333',
        strokeWidth: 0.6,
        dashStyle: null
    };

    mapLayer.renderingMode = renderingMode;
    mapLayer.isCanvasMode = mapLayer.layerContainer.isD3Selection !== true;
    mapLayer.renderingModeCapitialized = toCapitalize(renderingMode);
    mapLayer.paintLayer();
};

/**
 * for update repaint with the settings
 */
MapLayer.prototype.update = function(newLayer) {
    if (newLayer) {
        this.layer = newLayer;
    }
    this.paintLayer(true);
};

MapLayer.prototype.paintLayer = function(update) {
    var mapLayer = this;
    mapLayer[`paint${mapLayer.renderingModeCapitialized}Layer`](update); // init the current mode
};

/**
 *
 */
MapLayer.prototype.paintSvgLayer = function(update) {
    var mapLayer = this,
        areas = MapLayer.getMapFeatures(mapLayer.layer),
        applyBorderStyles = function(features) {
            features.attrs({
                stroke: function(feature) {
                    return mapLayer.getStyle(feature, 'strokeColor');
                },
                'stroke-width': function(feature) {
                    return mapLayer.getStyle(feature, 'strokeWidth');
                },
                'stroke-dasharray': function(feature) {
                    return mapLayer.getStyle(feature, 'dashStyle');
                },
                'stroke-opacity': function(feature) {
                    return mapLayer.getStyle(feature, 'strokeOpacity');
                }
            });
        },
        features = createElement('path', '.map-feature', MapLayer.getMapFeatures(areas), mapLayer.layerContainer).attrs({
            fill: function(feature) {
                return mapLayer.getStyle(feature, 'fillColor', true);
            },
            'fill-opacity': function(feature) {
                return mapLayer.getStyle(feature, 'fillOpacity') || 1;
            },
            d: mapLayer.utils.geoPath
        });

    if (mapLayer.useMesh && mapLayer.layer.length) {
        var mesh = mapLayer.geoMap.dataObject.getFeatureMesh(MapLayer.getMapFeatures(mapLayer.layer));
        mapLayer.mesh = features = createElement('path', '.map-feature-outline', [areas[0]], mapLayer.layerContainer).attrs({
            fill: 'none',
            d: mapLayer.utils.geoPath(mesh)
        });
        // make it to the top
        features.raise();

        applyBorderStyles(features);
    } else {
        if (mapLayer.mesh) {
            mapLayer.mesh.remove();
        }
        applyBorderStyles(features);
    }
};

/**
 *
 */
MapLayer.prototype.paintCanvasLayer = function(update) {
    var mapLayer = this,
        context = mapLayer.layerContainer,
        areas = MapLayer.getMapFeatures(mapLayer.layer);

    // update the context;
    mapLayer.utils.geoPath.context(context);

    each(areas, function(feature) {
        // set the styles
        context.save();
        context.fillStyle = mapLayer.getStyle(feature, 'fillColor', true);
        context.strokeStyle = mapLayer.getStyle(feature, 'strokeColor');
        context.lineWidth = mapLayer.getStyle(feature, 'strokeWidth');

        // set dash array only if user has given
        var dashArray = mapLayer.getStyle(feature, 'dashStyle');

        // #ZC1452
        if (dashArray && defined(context.setLineDash)) {
            context.setLineDash(dashArray);
        }

        context.beginPath();
        mapLayer.utils.geoPath(feature);

        //set the transform before the fill
        if (feature.transform) {
            context.translate(feature.transform.x, feature.transform.y);
        }

        context.fill();

        if (!mapLayer.useMesh) {
            context.stroke();
        }

        context.closePath();
        context.restore();
    });

    // #ZC1493 error when empty layer is passed.
    if (mapLayer.useMesh && areas.length) {
        var mesh = mapLayer.geoMap.dataObject.getFeatureMesh(MapLayer.getMapFeatures(mapLayer.layer));

        if (!mesh) {
            return;
        }

        context.save();
        context.fillStyle = mapLayer.getStyle(areas[0], 'fillColor', true);
        context.strokeStyle = mapLayer.getStyle(areas[0], 'strokeColor');
        context.lineWidth = mapLayer.getStyle(areas[0], 'strokeWidth');

        var dashArray = mapLayer.getStyle(mesh, 'dashStyle');

        // #ZC1452
        if (dashArray && defined(context.setLineDash)) {
            context.setLineDash(dashArray);
        }

        context.beginPath();
        mapLayer.utils.geoPath(mesh);
        context.stroke();
        context.closePath();
        context.restore();
    }
};

MapLayer.prototype.destroy = function() {
    var mapLayer = this;
    mapLayer.update([]); // remove all layer
    each(mapLayer, function(value, key) {
        mapLayer[key] = null;
        delete mapLayer[key];
    });
};
