// @zohocharts/zc/src/geo-core/map-data-parser.js
//map-data--parser.js
import { pick, isUndefined, simpleClone, isArray, defined } from '../utils/zcutils';
import { ORDINAL_DATATYPE, NUMERIC_DATATYPE } from '../core/datatype-constants';
import { getFormattedValue, getFormattedValueByPercent } from '../formatter/formatter';
import { STACKLABELS_LABELTYPE } from '../label/labeltype-constants';
import { colorscale_helpers } from '../axis/colorscale-helpers';
import { PER_SHOWAS, SPLITPER_SHOWAS, MAXPER_SHOWAS } from '../label/showas-constants';
import { encodeHTML, decodeHTML } from '../utils/zcutils';
import { CUSTOM_VALUE_PATTERN } from '../core/main-zc-init';
import { d3_set } from '../core/d3-constants';
import { getTotal } from '../core/data-parser';

/**
 *
 * @param {GeoMap} geoMap
 * @constructor
 */
export function MapDataSet(geoMap) {
    this.geoMap = geoMap;
    this.initialize();
}

/**
 * initialize the map data set.
 * It will create the relation between the columns and the series.
 *
 */
MapDataSet.prototype.initialize = function() {
    var dataSet = this,
        geoMap = dataSet.geoMap,
        userData = geoMap.userdata,
        metaData = userData.metadata,
        columns = (dataSet.columns = pick(metaData.columns, {})),
        axes = (dataSet.axes = pick(metaData.axes, { geo: [0], clr: [1], z: [1] })),
        colorColumn = pick(columns[axes.clr], {}),
        geoColumn = pick(columns[axes.geo], {}),
        zColumn = pick(columns[axes.z], {}),
        geoMapByKey;

    // set the color index in dataSet

    dataSet.latLongJoinByKeys = d3_set(['lat-long', 'long-lat']);
    dataSet.clridx = pick(colorColumn.dataindex);
    dataSet.zidx = pick(zColumn.dataindex);
    dataSet.geoidx = pick(geoColumn.dataindex, 0);
    dataSet.geoMapByKey = geoMapByKey = dataSet.getJoinByKey(geoColumn);

    if (axes.label && columns[axes.label]) {
        dataSet.labelIndex = columns[axes.label].dataindex;
    }

    if (dataSet.latLongJoinByKeys.has(dataSet.geoMapByKey)) {
        if (geoMapByKey === 'lat-long') {
            dataSet.latitudeIndex = 0;
            dataSet.longitudeIndex = 1;
        } else {
            dataSet.latitudeIndex = 1;
            dataSet.longitudeIndex = 0;
        }
    }
};

/**
 *
 * @param column
 */
MapDataSet.prototype.getJoinByKey = function(column) {
    var dataSet = this,
        geoColumn = pick(column, dataSet.columns[dataSet.axes.geo], {}),
        joinBy = pick(geoColumn.joinBy, 'zc-key');

    /**
     *
     * 1. This checks all joinBy key names except "zc-key". Since it is the standard key available
     *    in all geojson source.
     *
     * 2. If the joinBy key is defined by the user but the key is missing in the actual geojson properties,
     *    then it replaces the joinBy key with `zc-key` by default.
     */
    if (joinBy !== 'zc-key' && !dataSet.latLongJoinByKeys.has(joinBy)) {
        // take the first feature for testing assuming that remaining feature will have the properties.
        var feature = dataSet.geoMap.geo.defaultLayer.layerGeojson.features[0].properties;
        if (!defined(feature[joinBy])) {
            joinBy = 'zc-key';
        }
    }

    return joinBy;
};

/**
 *
 * @param type
 * @return {*}
 */
MapDataSet.prototype.getColumnIdx = function(type) {
    if (isArray(this.axes[type])) {
        return this.axes[type][0];
    }
    return null;
};

/**
 *
 * @param type
 * @return {"numeric"|"ordinal"|"time"}
 */
MapDataSet.prototype.getDataType = function(type) {
    var dataSet = this,
        columnIndex = dataSet.getColumnIdx(type);

    if (defined(dataSet.columns[columnIndex])) {
        return dataSet.columns[columnIndex].datatype;
    }
    return ORDINAL_DATATYPE;
};

/**
 *
 * @param point
 * @return {*}
 */
MapDataSet.prototype.getZ = function(point) {
    return point[this.zidx];
};

/**
 *
 * @param point
 * @return {*}
 */
MapDataSet.prototype.getClr = function(point) {
    return point[this.clridx];
};

/**
 *
 * @param point
 * @return {*}
 */
MapDataSet.prototype.getGeo = function(point) {
    return point[this.geoidx];
};

/**
 *
 * @param point
 * @return {number}
 */
MapDataSet.prototype.getLatitude = function(point) {
    return point[this.latitudeIndex];
};

/**
 *
 * @param point
 * @return {number}
 */
MapDataSet.prototype.getLongitude = function(point) {
    return point[this.longitudeIndex];
};

/**
 *
 * @return {any}
 */
MapDataSet.prototype.getMapByKey = function() {
    return this.geoMapByKey;
};

/**
 *
 * @param type
 * @return {*}
 */
MapDataSet.prototype.getColumn = function(type) {
    return this.columns[this.getColumnIdx(type)];
};

MapDataSet.prototype.getX = MapDataSet.prototype.getY = MapDataSet.prototype.getGeo;

/**
 *
 * @param chartObj
 * @param dataPoint
 * @param colidx
 * @param seriespos
 * @param percent
 * @param type
 * @param subseriespos
 * @param toDecode
 * @param subseriesLength
 * @return {*}
 */
MapDataSet.prototype.getDataPointValue = function(
    chartObj,
    dataPoint,
    colidx,
    seriespos,
    percent,
    type,
    subseriespos,
    toDecode,
    subseriesLength
) {
    if (colidx) {
        colidx = +colidx;
    }

    if (isUndefined(toDecode)) {
        toDecode = !/(tooltip|legend)/.test(type);
    }

    var axes = this.axes,
        columns = this.columns,
        colInfo = columns[colidx] || {},
        datatype = colInfo.datatype,
        dataindex = colInfo.dataindex;

    var format = colInfo[datatype];

    if (dataPoint != null) {
        //Even though clrIdx is not specified in metadata, the seriesname will be returned correctly
        if ((axes.clr != null || type === 'legend') && colidx == axes.clr) {
            if (defined(this.clridx) && colorscale_helpers.isQuantileColorScale(chartObj)) {
                if (type === 'legend') {
                    return getFormattedValue(format, dataPoint, datatype, chartObj, type, toDecode, '-');
                }
                return getFormattedValue(format, this.getClr(dataPoint), datatype, chartObj, type, toDecode);
            } else {
                return getFormattedValue(format, this.geoMap.seriesdata[seriespos].seriesname, datatype, chartObj, type, toDecode);
            }
        } else {
            var value;
            var formatFn = percent ? getFormattedValueByPercent : getFormattedValue;
            if (subseriesLength) {
                //subserieslength - to iterate all subseries y value to show for tooltip shared series info
                var cumval = 0;
                for (var j = 0; j < subseriesLength; j++) {
                    cumval += this.getDataPointUnformattedValue(datatype, seriespos, j, colidx, dataPoint, type, percent, dataindex);
                }
                value = cumval;
            } else {
                value = this.getDataPointUnformattedValue(datatype, seriespos, subseriespos, colidx, dataPoint, type, percent, dataindex);
            }
            if (value === '' && percent) {
                return value;
            }
            return formatFn(format, value, datatype, chartObj, type, toDecode);
        }
    }
    return null;
};

/**
 *
 * @param dataType
 * @param seriesIndex
 * @param subSeriesIndex
 * @param columnIndex
 * @param dataPoint
 * @param type
 * @param percent
 * @param dataIndex
 * @return {*}
 */
MapDataSet.prototype.getDataPointUnformattedValue = function(
    dataType,
    seriesIndex,
    subSeriesIndex,
    columnIndex,
    dataPoint,
    type,
    percent,
    dataIndex
) {
    var geoMap = this.geoMap,
        value = '';

    if (percent) {
        var total;

        if (dataType === NUMERIC_DATATYPE) {
            total = getTotal(geoMap, seriesIndex, subSeriesIndex, columnIndex, dataPoint, percent, type);
            if (defined(total)) {
                value = pick(((type === STACKLABELS_LABELTYPE ? dataPoint : dataPoint[dataIndex]) / total) * 100, 0);
                return value;
            }
        }
    }
    return type === STACKLABELS_LABELTYPE ? dataPoint : dataPoint[dataIndex];
};

/**
 *
 * @param geoMap
 * @param columnIndex
 * @param seriesIndex
 * @param subSeriesIndex
 * @param itemIndex
 * @param point
 * @return {*}
 */
MapDataSet.prototype.getDataPoint = function(geoMap, columnIndex, seriesIndex, subSeriesIndex, itemIndex, point) {
    if (defined(point)) {
        return point;
    }

    var series = geoMap.seriesdata[seriesIndex].data,
        seriesAfterUpdate = geoMap.renderer[seriesIndex].seriesdataAfterUpdate;

    if (seriesAfterUpdate) {
        return seriesAfterUpdate[subSeriesIndex][itemIndex];
    } else {
        return series[subSeriesIndex][itemIndex];
    }
};

/**
 *
 *
 * @param chartObj
 * @param colidx
 * @param seriespos
 * @param subseriespos
 * @param itempos
 * @param d
 * @param type
 * @param color
 * @param toDecode
 * @param subseriesLength
 * @return {string}
 */
MapDataSet.prototype.getCustomizedVal = function(
    chartObj,
    colidx,
    seriespos,
    subseriespos,
    itempos,
    d,
    type,
    color,
    toDecode,
    subseriesLength
) {
    var dataSet = this;

    if (isUndefined(toDecode)) {
        toDecode = !/(tooltip|legend)/.test(type);
    }

    return colidx.toString().replace(CUSTOM_VALUE_PATTERN, function() {
        var matchedString = arguments[0],
            valIndex = arguments[1], //INDEX value of {{val(INDEX)}}
            colIndex = arguments[2], //INDEX value of {{col(INDEX)}}
            perIndex = arguments[3], //INDEX value of {{per(INDEX)}}
            splitPerIndex = arguments[4], //INDEX value of {{splitper(INDEX)}}
            maxPerIndex = arguments[5],
            geoProperty = arguments[6],
            idx = valIndex || colIndex || perIndex || splitPerIndex || maxPerIndex,
            dataPoint;

        if (matchedString.indexOf('{{clr}}') !== -1) {
            // theme color based on series color
            return color;
        }

        if (dataSet.columns[idx] != null) {
            dataPoint = type === STACKLABELS_LABELTYPE ? d : dataSet.getDataPoint(chartObj, idx, seriespos, subseriespos, itempos, d);
            if (dataPoint != null) {
                var operation = matchedString.match(/[a-z]+/gi)[0];
                switch (operation) {
                    case 'col':
                        return dataSet.getColumnName(colIndex, toDecode); //return columnname
                    case 'val':
                        return dataSet.getDataPointValue(
                            chartObj,
                            dataPoint,
                            valIndex,
                            seriespos,
                            false,
                            type,
                            subseriespos,
                            toDecode,
                            subseriesLength
                        ); //return value
                    case PER_SHOWAS:
                    case SPLITPER_SHOWAS:
                    case MAXPER_SHOWAS:
                        return dataSet.getDataPointValue(
                            chartObj,
                            dataPoint,
                            perIndex || splitPerIndex || maxPerIndex,
                            seriespos,
                            operation,
                            type,
                            subseriespos,
                            toDecode,
                            subseriesLength
                        );
                }
            } else {
                return '';
            }
        } else if (geoProperty) {
            // #ZC1469
            dataPoint = dataSet.getDataPoint(chartObj, idx, seriespos, subseriespos, itempos, d);
            var feature = chartObj.dataObject.getMapFeaturesByDataPoint(dataPoint, 'array');
            if (feature.length) {
                return feature[0].properties[geoProperty] || matchedString;
            }
        } else {
            return matchedString;
        }
    });
};

MapDataSet.prototype.getColumnName = function(colidx, toDecode) {
    var dataSet = this;

    if (toDecode) {
        return encodeHTML(dataSet.columns[colidx].columnname);
    } else {
        return decodeHTML(dataSet.columns[colidx].columnname);
    }
};

/**
 *
 * @param chartObj
 * @param data
 * @param seriesIndex
 * @return {Array}
 */
MapDataSet.prototype.getMetaInfoObj = function(chartObj, data, seriesIndex) {
    var dataSet = this,
        axes = dataSet.axes,
        columns = dataSet.columns,
        seriesdata = dataSet.geoMap.seriesdata,
        dataWithMetadata = [],
        getColObj = function(columnIndex, value) {
            var obj = simpleClone(columns[columnIndex] || {});
            obj.value = value;
            obj.columnindex = columnIndex;
            return obj;
        };

    if (isArray(data)) {
        data.forEach(function(d, i) {
            var columnIndex = columns.indexOf(
                columns.filter(function(c, j) {
                    return c.dataindex === i;
                })[0]
            );
            dataWithMetadata.push(getColObj(columnIndex, d));
        });
    }

    if (defined(axes.clr)) {
        var clrColIdx = isArray(axes.clr) ? axes.clr[0] : axes.clr;
        dataWithMetadata.push(getColObj(clrColIdx, seriesdata[seriesIndex].seriesname));
    }

    return dataWithMetadata;
};
