// @zohocharts/zc/src/geo-core/mapOptions.js
import { defaultTheme as options } from '../core/defaulttheme';

options.map = function() {
    return {
        allAreas: true,
        renderingMode: window.rendererMode || 'svg',
        fillColor: '#dddddd',
        strokeColor: '#444',
        strokeWidth: 0.6,
        gradients: {
            options: {
                linear: {
                    y2: 100
                },
                radial: {
                    radius: 100,
                    gradientUnits: 'objectBoundingBox'
                }
            }
        },
        globe: {
            strokeWidth: 1.5
        },
        labels: {
            strokeWidth: 1.5
        },
        extent: {
            top: 10,
            right: 10,
            bottom: 10,
            left: 10
        },
        layerName: 'source',
        useMeshOutline: true,
        textStrokeWidth: 2,
        animation: {
            enabled: true,
            type: 'fade',
            duration: 500,
            easingType: 'cubic'
        }
    };
};

export var _mapOptions = options.map;
