// @zohocharts/zc/src/geo-core/zc_map_events.js
import { ZC_ChartEvents } from '../events/chart-events';
import { GeoPoint } from './coordinates/index';
import { getPlotDefaultBehaviour } from '../events/visualEffects/behaviour-mapping';
import {d3_merge} from "../core/d3-constants";
import {Quadtreesearch} from "../events/dataSelection/tools/quadtree-search";

export class ZC_MapEvents extends ZC_ChartEvents {
    init() {
        super.init();
        this.addListeners(this.tracker);

        this.eventManager.dataFinder({
            mousemove: ZC_MapEvents.findGeoData,
            click: ZC_MapEvents.findGeoData,
            doubleclick: ZC_MapEvents.findGeoData
        });

        this.eventManager.defaultBehaviour = getPlotDefaultBehaviour();
        this.eventManager.customBehaviour = ZC_ChartEvents.getCustomBehaviour;
    }

    addListeners(tracker, type) {
        super.addListeners(tracker, type);
    }

    static findGeoData(mouse) {
        let geoMap = this.chart,
            mapEvents = this.chart.eventHandler.mapEvents,
            geoPoint = new GeoPoint(mouse.x, mouse.y),
            pointObj = mapEvents.getPoint(geoPoint),
            point = pointObj ? pointObj.point : null;
        if (point) {
            pointObj.seriesIndex = point.seriesIndex;
            pointObj.itemIndex = point.itemIndex;
        }

        return {
            nearest: pointObj
        };
    }
}

export function getGeoEnclosedPoints(data, key, mouse, mode, chart){
    let accessor = chart.eventHandler.dataAccesor(mouse),
        searchCache = chart.cache.search,
        enclosed,
        tree;

    if (mode == 'QUADTREE') {
        let d = data,
            radius = accessor.radius(key),
            sizeOffset = accessor.sizeOffset(key),
            extent = accessor.extent(),
            size = accessor.size(key),
            pane = '0_0';
        tree = searchCache[`${pane}_${key}_${mode}`] || new Quadtreesearch(d, size, radius, extent, sizeOffset);
        enclosed = tree.find(mouse);
    }

    return enclosed;
}
