// @zohocharts/zc/src/sparkline/canvasSparkline.js
//$Id$
//canvasSparkline.js

import { getAxisRotated } from "../axis/axisutils";
import { getOrdinalRange } from "../axis/scale";
import { $Browser } from "../core/browser-detect";
import { DataProcessor, getChartName } from "../core/chart-constructor";
import { d3_event, d3_select } from "../core/d3-constants";
import { NUMERIC_DATATYPE, ORDINAL_DATATYPE } from "../core/datatype-constants";
import { VizLayout } from "../core/main-constructor";
import { plot, renderer_helpers, scale, canvas_plot_renderer } from "../core/main-zc-init";
import { PX, mathFloor } from "../core/zc-constants";
import { setGlobalPropByUserdata } from "../core/zcglobal";
import { stateChangeListener } from "../events/eventActions";
import { appendEle } from "../utils/zcsvgutils";
import { destroyObjectProperties, fixHDPI, getPropVal, invokeFunction, isObjectPropDefined } from "../utils/zcutils";
import { __defaultSparklineDataFinder, __defaultSparklinePlotHandlers } from "./sparklineEvents";

canvas_plot_renderer.sparkline = function (chart) {
    const sparkline = chart.sparkline = new Sparkline(chart);

    try {
        sparkline.declareProperties();
        sparkline.prepareLayout();
        sparkline.processUserData();
        sparkline.constructScales();
    } catch (err) {
        VizLayout.helpers.handleErrors(chart, err);
    }
}

function Sparkline(chart) {
    this.chart = chart;
    this.container = chart.container;
    this.userdata = chart.userdata;
}

Sparkline.prototype.declareProperties = function () {
    const chart = this.chart;
    chart.cache = {};
    chart.plot = {};
    chart.dataset = {};
    chart.dataObject = {};
    chart.timerObj = [];
    chart.datalabels = {};
    chart.legendarea = {};
}

Sparkline.prototype.prepareLayout = function (update) {
    const chart = this.chart;
    const container = this.container;
    const width = chart.chartWidth = chart.getChartDimension('width');
    const height = chart.chartHeight = chart.getChartDimension('height');
    chart.setCanvasDimension(width, height);

    if (!update) {
        container.selectAll('*').remove(); // cleans container
        container.styles({
            position: "relative"
        })

        appendEle(chart.container, 'canvas', [1], '.', 'class', 'zc-canvas'); // adding zc-canvas
    }
    chart.context = this.prepareCanvas();
}

Sparkline.prototype.prepareCanvas = function () {
    const chart = this.chart;
    const canvas = this.container.select('.zc-canvas');
    const width = chart.getChartDimension('width');
    const height = chart.getChartDimension('height');
    const { marginLeft = 10, marginRight = 10, marginTop = 10, marginBottom = 10 } = getPropVal(chart, 'userdata.chart');

    canvas.attrs({
        width: width - marginRight,
        height: height - marginBottom
    })
    canvas.styles({
        position: "absolute",
        // "pointer-events": "none",
        left: marginLeft + PX,
        top: marginTop + PX,
        width: (width - marginRight) + PX,
        height: (height - marginBottom) + PX,
        visibility: "visible",
        opacity: 1
    });

    const context = canvas.node().getContext('2d');
    const imageSmoothingEnabled = `${$Browser.prefix.lowercase}ImageSmoothingEnabled`;

    context.clearRect(0, 0, width, height);
    context[imageSmoothingEnabled] = false; // fix old browser

    return fixHDPI(context);
}

Sparkline.prototype.processUserData = function () {
    const chart = this.chart;
    const data = this.userdata;
    const width = chart.getChartDimension('width');
    const height = chart.getChartDimension('height');

    chart.chartTypes = DataProcessor.helpers.getAllChartTypes(data, 'object');
    chart.globalChartType = data.seriesdata && data.seriesdata.type;
    chart.chartFontProp = {};
    chart.globalProp = setGlobalPropByUserdata(chart.userdata, { width, height }, chart.chartFontProp);
    chart.themeoptions = chart.getThemeOptions(chart.globalProp);

    chart.processUserData(chart.themeoptions);
    this.updateSystemConf();

    chart.setChartDimension();
    chart.updatePlotArea();

    chart.getChartSeriesColor();
    chart.isAxisCategory = chart.dataObject.isAxisCategory;
    chart.isNonAxisMultiSeriesCategory = chart.dataObject.isNonAxisMultiSeriesCategory;
    chart.isPolarAxisCategory = chart.dataObject.isPolarAxisCategory;
}

Sparkline.prototype.updateSystemConf = function () {
    const chart = this.chart;
    const systemConf = chart.systemConf;
    const plot = systemConf.chart.plot;
    const plotoptions = plot.plotoptions;

    plot.datalabels.show = false;
    systemConf.chart.scroll ? (systemConf.chart.scroll.enabled = false) : null;
    systemConf.loader.enabled = false;

    for (const key in plotoptions) {
        const datalabels = plotoptions[key].datalabels;
        datalabels ? (datalabels.show = false) : null;
    }
    // const chartOptions = userdata.chart || {};
    // systemConf.chart.marginLeft = chartOptions.marginLeft || 0;
    // systemConf.chart.marginRight = chartOptions.marginRight || 0;
    // systemConf.chart.marginTop = chartOptions.marginTop || 0;
    // systemConf.chart.marginBottom = chartOptions.marginBottom || 0;
}

Sparkline.prototype.constructScales = function () {
    const chart = this.chart;
    const systemConf = chart.systemConf;
    const axes = chart.axes;
    const xaxis = axes.x;
    const yaxis = axes.y;

    if (chart._isAxisCategory() || chart.dataObject.isPolarAxisCategory || chart.dataObject.hasLevelMarker) {
        axes.rotated = getAxisRotated(systemConf);
        [xaxis.orient, yaxis.orient] = axes.rotated ? ['left', 'bottom'] : ['bottom', 'left'];
        chart.ismultiy = chart.getAxesCount('y') > 1;
        chart.ismultix = chart.getAxesCount('x') > 1;
    }

    scale.constructor(chart, 'bottom');     // x-axis
    scale.constructor(chart, 'left');       // y-axis
    this.addScaleDependencies();
}

Sparkline.prototype.addScaleDependencies = function () {
    const chart = this.chart;
    const xscale = chart.axes.x.scale;
    const ygroups = chart.axes.y.groups;
    const domain = xscale.domain(),
        range = xscale.range(),
        outerPadding = xscale.outerPaddingPx,
        interval = (range[1] - range[0]) / domain.length;

    xscale.invert = xscale.invert || function (_) {
        return domain[mathFloor((_ - range[0]) / interval)];
    }
    xscale.index = xscale.index || function (_) {
        return mathFloor((_ - range[0]) / interval);
    }

    if (chart.dataset.getDataType('x') == ORDINAL_DATATYPE) {
        scale.barBandwidth(chart);
        ygroups.forEach((ygroup) => getOrdinalRange(ygroup.scale));
    }

}

Sparkline.prototype.draw = function (data) {
    const chart = this.chart;

    this.render(false);
}

Sparkline.prototype.render = function (update) {
    const chart = this.chart;
    const sparkline = this;
    const systemConf = chart.systemConf;
    const renderer = chart.renderer = chart.renderer || [];
    const rendererConf = chart.rendererConf = chart.rendererConf || [];
    const seriesdata = chart.seriesdata;
    const dp_helpers = DataProcessor.helpers;

    chart.plot.renderer = { mode: "canvas" };

    try {
        seriesdata.forEach(function (series, index) {
            const type = dp_helpers.getChartType(series.type || chart.globalChartType, 'string');
            const type_num = dp_helpers.getChartType(series.type || chart.globalChartType, NUMERIC_DATATYPE);
            const chartName = getChartName(type);
            renderer[index] = renderer[index] || new canvas_plot_renderer[chartName]();
            rendererConf[index] = new plot.rendererConf();
            renderer[index].chartRendered = false;

            sparkline.addDependencies(type);

            const args = {
                barDatasetIndex: 0,
                labelArray: {},
                labelQueue: [],
                notesArray: []
            }

            renderer_helpers.setRendererProp(chart, renderer[index], series, index, index, args, index, type_num, 0);
        })

        const plotarea = chart.plotarea;
        seriesdata.forEach(function (series, index) {
            if (plotarea.width > 0 && plotarea.height > 0) {
                const { bound } = chart.renderer[index].commonRendererProp().params;
                const { x, y } = bound;
                chart.context.translate(x, y);
                renderer[index](chart.context, series);
            }
        })

        this.handleEvents(update);

        // if (chart.plot.renderer) {
        //     chart.plot.renderer.draw(update);
        // } else {
        //     var renderer = (chart.plot.renderer = plot.render(chart, update));
        //     renderer.draw(false);
        // }
        chart.chartRendered = true;

        stateChangeListener(chart, systemConf.canvas.events, 'chart', update ? 'onreload' : 'onload');
    } catch (err) {
        VizLayout.helpers.handleErrors(chart, err);
    }
}

// Renderer dependencies
Sparkline.prototype.addDependencies = function (chartType) {
    const chart = this.chart;
    if (chart.dataset.getDataType('x') == ORDINAL_DATATYPE) {
        scale.barBandwidth(chart);
    }
}

Sparkline.prototype.handleEvents = function (update) {
    const chart = this.chart;
    this.eventManager = {};
    const canvas = d3_select(chart.context.canvas);
    chart.trackerElement = canvas;
    canvas.styles({
        "pointer-events": "all"
    })

    this.bindEvents(update);
}

Sparkline.prototype.bindEvents = function (update) {
    const _this = this;
    const chart = this.chart;
    const element = chart.container;
    const tracker = chart.trackerElement;
    const handlers = __defaultSparklinePlotHandlers
    const events = {
        click: function (args) {  //clickHandlers
            handlers.click.call(_this, args)
        },
        dblclick: function (args) {
            handlers.doubleclick.call(_this, args)
        },
        contextmenu: function (args) {
            handlers.contextmenu.call(_this, args)
        },
        mousemove: function (args) {  //moveHandlers
            handlers.mousemove.call(_this, args)
        },
        mouseout: function (args) {
            handlers.mouseout.call(_this, args)
        },
        tap: function (args) {    //touchHandlers
            handlers.click.call(_this, args)
        },
        dbltap: function (args) {
            handlers.doubleclick.call(_this, args)
        },
        singletouchmove: function (args) {
            handlers.mousemove.call(_this, args)
        },
        touchend: function (args) {
            handlers.mouseout.call(_this, args)
        }
    };

    for (const eventName in events) {
        tracker.on(eventName, events[eventName]);
    }

}

Sparkline.prototype.destroy = async function () {
    const chart = this.chart,
        sparkline = chart.sparkline;
    if (chart.destroying) {
        return;
    }
    return new Promise(function () {
        chart.destroying = true;

        invokeFunction(isObjectPropDefined(chart.systemConf, 'canvas.events.destroy', true), d3_event(), chart);

        // remove the container reference
        sparkline.container.datum(null);
        chart.container = chart.sparkline.container = null;


        if (chart.dataObject && (chart.dataObject.isAxisCategory || chart.dataObject.isPolarAxisCategory)) {
            // remove axis
            destroyObjectProperties(chart.axes.x);
            chart.axes.y.groups.forEach(yAxis => destroyObjectProperties(yAxis));
            // remove label array #ZC3245
            chart.renderer.forEach(renderer => destroyObjectProperties(renderer.commonRendererProp().params.labelArray));

            // remove all renderer
            chart.renderer.forEach(renderer => destroyObjectProperties(renderer, chart));
            destroyObjectProperties(chart.dataObject, chart);
            destroyObjectProperties(chart.cache);
            destroyObjectProperties(chart);

            // resolve(true);
        }
    })

}

Sparkline.prototype.redraw = function () {
    const chart = this.chart;
    chart.updating = true;
    chart.phase = 'redraw';
    chart.viewPortExtent = null;

    chart._chartwidth = null;
    chart._chartheight = null;

    chart.oldChartTypes = chart.chartTypes;
    chart.chartRendered = false;

    this.prepareLayout(true);
    this.processUserData();
    this.constructScales();
    this.render(true);

    chart.phase = null;
}

export var _SparklineChart = canvas_plot_renderer.sparkline;
