// @zohocharts/zc/src/sparkline/sparklineEvents.js
import { d3_event } from "../core/d3-constants";
import { d3_mouse } from "../core/d3/mouse";
import { getPropVal, invokeFunction } from "../utils/zcutils";

export const __defaultSparklinePlotHandlers = {
    mousemove: SPARKLINE_MOUSEMOVE,
    mouseout: SPARKLINE_MOUSEOUT,
    click: SPARKLINE_CLICK,
    doubleclick: SPARKLINE_DOUBLECLICK,
    contextmenu: SPARKLINE_CONTEXTMENU,
    tap: SPARKLINE_TAP,
    dbltap: SPARKLINE_DBLTAP,
    singletouchmove: SPARKLINE_SINGLETOUCHMOVE,
    touchend: SPARKLINE_TOUCHEND
}

export const __defaultSparklineDataFinder = {
    mousemove: dataFinder,
    click: dataFinder,
    doubleclick: dataFinder
}

function dataFinder(mouse) {
    const manager = this;
    const chart = manager.chart;
    const xScale = chart.axes.x.scale;
    const point = xScale.invert(mouse[0]);
    const index = xScale.index(mouse[0]);

    return {
        // nearest: point,
        // nearby: point,
        itemIndex: index,
        // dependents: null,
        // surround: null,
        mouse: mouse
    }
}

function SPARKLINE_MOUSEMOVE(event, d) {
    d3_event(event);
    const self = this,
        chart = this.chart,
        eventManager = this.eventManager,
        mouse = d3_mouse(chart.trackerElement);

    const point = dataFinder.call(this, mouse);

    if (chart.chartRendered) {
        eventManager.eventType =  'mousemove';
        eventManager.targetEvent = event;

        chart.rendererConf.forEach(conf => {
            const customEvent = getPropVal(conf.params.commonPlotOptions.events, eventManager.eventType);
            invokeFunction(customEvent, eventManager.targetEvent, point, chart);
        })
    }
}

function SPARKLINE_MOUSEOUT(event, d) {
    d3_event(event);
    const self = this,
        chart = this.chart,
        eventManager = this.eventManager,
        mouse = d3_mouse(chart.trackerElement);

    const point = dataFinder.call(this, mouse);

    if (chart.chartRendered) {
        eventManager.eventType =  'mouseout';
        eventManager.targetEvent = event;

        chart.rendererConf.forEach(conf => {
            const customEvent = getPropVal(conf.params.commonPlotOptions.events, eventManager.eventType);
            invokeFunction(customEvent, eventManager.targetEvent, point, chart);
        })
    }
}

function SPARKLINE_CLICK(event, d) {
    d3_event(event);
    const self = this,
        chart = this.chart,
        eventManager = this.eventManager,
        mouse = d3_mouse(chart.trackerElement);

    const point = dataFinder.call(this, mouse);

    if (chart.chartRendered) {
        eventManager.eventType =  'click';
        eventManager.targetEvent = event;

        chart.rendererConf.forEach(conf => {
            const customEvent = getPropVal(conf.params.commonPlotOptions.events, eventManager.eventType);
            invokeFunction(customEvent, eventManager.targetEvent, point, chart);
        })
    }
}

function SPARKLINE_DOUBLECLICK(event, d) {
    d3_event(event);
    const self = this,
        chart = this.chart,
        eventManager = this.eventManager,
        mouse = d3_mouse(chart.trackerElement);

    const point = dataFinder.call(this, mouse);

    if (chart.chartRendered) {
        eventManager.eventType =  'doubleclick';
        eventManager.targetEvent = event;

        chart.rendererConf.forEach(conf => {
            const customEvent = getPropVal(conf.params.commonPlotOptions.events, eventManager.eventType);
            invokeFunction(customEvent, eventManager.targetEvent, point, chart);
        })
    }
}

function SPARKLINE_CONTEXTMENU(event, d) {
    d3_event(event);
    const self = this,
        chart = this.chart,
        eventManager = this.eventManager,
        mouse = d3_mouse(chart.trackerElement);

    const point = dataFinder.call(this, mouse);

    if (chart.chartRendered) {
        eventManager.eventType =  'contextmenu';
        eventManager.targetEvent = event;

        chart.rendererConf.forEach(conf => {
            const customEvent = getPropVal(conf.params.commonPlotOptions.events, eventManager.eventType);
            invokeFunction(customEvent, eventManager.targetEvent, point, chart);
        })
    }
}

function SPARKLINE_TAP(event, d) {
    d3_event(event);
    const self = this,
        chart = this.chart,
        eventManager = this.eventManager,
        mouse = d3_mouse(chart.trackerElement);

    const point = dataFinder.call(this, mouse);

    if (chart.chartRendered) {
        eventManager.eventType =  'tap';
        eventManager.targetEvent = event;

        chart.rendererConf.forEach(conf => {
            const customEvent = getPropVal(conf.params.commonPlotOptions.events, eventManager.eventType);
            invokeFunction(customEvent, eventManager.targetEvent, point, chart);
        })
    }
}

function SPARKLINE_DBLTAP(event, d) {
    d3_event(event);
    const self = this,
        chart = this.chart,
        eventManager = this.eventManager,
        mouse = d3_mouse(chart.trackerElement);

    const point = dataFinder.call(this, mouse);

    if (chart.chartRendered) {
        eventManager.eventType =  'dbltap';
        eventManager.targetEvent = event;

        chart.rendererConf.forEach(conf => {
            const customEvent = getPropVal(conf.params.commonPlotOptions.events, eventManager.eventType);
            invokeFunction(customEvent, eventManager.targetEvent, point, chart);
        })
    }
}

function SPARKLINE_SINGLETOUCHMOVE(event, d) {
    d3_event(event);
    const self = this,
        chart = this.chart,
        eventManager = this.eventManager,
        mouse = d3_mouse(chart.trackerElement);

    const point = dataFinder.call(this, mouse);

    if (chart.chartRendered) {
        eventManager.eventType =  'singletouchmove';
        eventManager.targetEvent = event;

        chart.rendererConf.forEach(conf => {
            const customEvent = getPropVal(conf.params.commonPlotOptions.events, eventManager.eventType);
            invokeFunction(customEvent, eventManager.targetEvent, point, chart);
        })
    }
}

function SPARKLINE_TOUCHEND(event, d) {
    d3_event(event);
    const self = this,
        chart = this.chart,
        eventManager = this.eventManager,
        mouse = d3_mouse(chart.trackerElement);

    const point = dataFinder.call(this, mouse);

    if (chart.chartRendered) {
        eventManager.eventType =  'touchend';
        eventManager.targetEvent = event;

        chart.rendererConf.forEach(conf => {
            const customEvent = getPropVal(conf.params.commonPlotOptions.events, eventManager.eventType);
            invokeFunction(customEvent, eventManager.targetEvent, point, chart);
        })
    }
}
