// @zohocharts/zc/src/events/dataSelection/nearest-data.js
import { getSearchMode, getFineSearchKey, getChartModel } from './search-mode';
import { Quadtreesearch } from './tools/quadtree-search';
import { PolySearch } from './tools/poly-search';
import { Bisectsearch } from './tools/bisect-search';
import { getSelectionTool, SELECTION_TOOL } from './selection-tool';
import { isArray, isObject, objectToArray, invokeFunction } from '../../utils/zcutils';
import { d3_merge } from '../../core/d3-constants';
import { PathSearch } from './tools/path-search';
import { isDependentPointHighlightType } from '../events-utils';
import { isPolarSupportedChart } from '../../core/charttype-helpers';

export function getNearestPoint(data, mouse, isDefaultBisect, accessor) {
    //data contains array of objects with different keys of different series data
    let keys = data.keys(),
        //Group output points by similar search mode
        selected = {},
        nearby = [],
        dependents;
    keys.map(key => {
        //Get object by key. Here key refers different chart types / marks
        let d = data.get(key),
            super_key = d.superKey || key,
            axis_key = isPolarSupportedChart(accessor.chart, key) ? 'polar_' : '',
            bisect = accessor.isBisectMode(key, isDefaultBisect),
            // Group different series of data under same mark into array
            a = objectToArray(
                d,
                d0 => {
                    return !d0.disabled();
                },
                accessor.spliceIndices()
            ),
            //Find search tool
            mode = getSearchMode(axis_key + super_key, bisect),
            //Nearby points surrounded by radius from mouse
            encloseby = getEnclosedPoints(a, key, mouse, mode, accessor);

        //From enclosed array of points, find single
        //closest point by using respective selection tools
        if (encloseby) {
            nearby = nearby.concat(encloseby);

            let s = getSelectionTool(axis_key + super_key),
                //get closest single point
                p = s(encloseby, mouse, accessor.chart, key),
                //Find moder irrespective of shared criteria
                m = getSearchMode(super_key, false);
            if (p && ((isArray(p) && p.length) || isObject(p))) {
                if (selected[m]) {
                    selected[m].push(p);
                } else {
                    selected[m] = [p];
                }
            }
        }
    });

    let p0 = getNearestPointBySameMode(selected, mouse, accessor),
        nearest = getNearestPointByPriority(p0, accessor);

    if (nearest) {
        nearest = accessor.pointLiesInsidePlot(nearest) ? nearest : null;
    }
    // if the point is dependant type
    if (nearest && accessor.hasDependantPoints()) {
        dependents = getDependantPoints(nearest, accessor);
    }

    let point = { nearest, nearby, dependents },
        { userSelection } = accessor;

    if (userSelection) {
        point = invokeFunction(userSelection, point, accessor.chart) || point;
    }

    return point;
}

/**
 *
 * @param point
 * @param accessor
 * @return {*[]}
 */
function getDependantPoints(point, accessor) {
    let type = getPointType(point, accessor),
        chartModel = getChartModel(type),
        { seriesIndex } = point,
        renderer = accessor.chart.renderer[seriesIndex],
        points;

    if (chartModel === 'FLOW') {
        points = renderer.highlighter.getDependentLayoutPoints(point, true).map(d => d.point);
    }

    return points;
}

function getEnclosedPoints(data, key, mouse, mode, accessor) {
    let searchCache = accessor.chart.cache.search,
        enclosed,
        tree,
        pane = searchCache.lastPane;

    if (mode == 'QUADTREE') {
        //For Quad tree
        //For quad mode, we need to merge all series of data if its in array form
        let d = data[0] instanceof Array ? d3_merge(data) : data,
            radius = accessor.radius(key),
            sizeOffset = accessor.sizeOffset(key),
            extent = accessor.extent(),
            size = accessor.size(key);
        mouse = accessor.toArray(mouse);
        tree = searchCache[`${pane}_${key}_${mode}`] || new Quadtreesearch(d, size, radius, extent, sizeOffset);
        enclosed = tree.find(mouse);
    } else if (mode == 'POLY') {
        //For area charts and polygon highlight
        let d = d3_merge(data);
        tree = searchCache[`${pane}_${key}_${mode}`] || new PolySearch(d);
        enclosed = tree.find(mouse);
    } else if (mode === 'PATH') {
        //For area charts and polygon highlight
        let d = d3_merge(data);
        tree = searchCache[`${pane}_${key}_${mode}`] || new PathSearch(d);
        enclosed = tree.find(mouse);
    } else {
        //For bisecting data by x scale
        let m = accessor.bisectMode(),
            median = accessor.median();
        tree = searchCache[`${pane}_${key}_${mode}`] || new Bisectsearch(data, m);
        enclosed = tree.find(median, mouse);
    }
    searchCache[`${pane}_${key}_${mode}`] = tree;
    return enclosed;
}

function getNearestPointBySameMode(data, mouse, accessor) {
    var p = [];
    for (let mode in data) {
        let d = data[mode];
        if (d.length > 1) {
            let s = SELECTION_TOOL.get(`${mode}_TYPES`),
                key = getFineSearchKey(mode),
                d0 = s(d, mouse, accessor.chart, key, true);
            p.push(d0);
        } else {
            p.push(d[0]);
        }
    }
    return p;
}

function getNearestPointByPriority(data, accessor) {
    if (data && data.length == 1) {
        return data[0];
    } else {
        var high,
            rank = {
                QUADTREE: () => 0,
                BISECT: () => 1,
                POLY: () => 2,
                BISECT_QUADTREE: function(point) {
                    //changed while applying bar trigger offset
                    if (accessor.hasSelection(point, 'BISECT_QUADTREE')) {
                        return 1;
                    } else {
                        return 3;
                    }
                }
            },
            pointRankType = getPointRankType(data, accessor),
            sorted;

        if (pointRankType === 'SEARCH_MODE') {
            sorted = data.sort((p0, p1) => {
                if (p0 && ((isArray(p0) && p0.length) || isObject(p0))) {
                    var type_p0 = getPointType(p0, accessor),
                        mode_p0 = getSearchMode(type_p0),
                        type_p1 = getPointType(p1, accessor),
                        mode_p1 = getSearchMode(type_p1);
                    return rank[mode_p0](p0) - rank[mode_p1](p1);
                }
            });
        } else if (pointRankType === 'POINT') {
            sorted = data.sort((p1, p2) => {
                return p1._getRank() - p2._getRank();
            });
        }

        return sorted[0];
    }
}

function getPointType(p, accessor) {
    var seriesindex = p.seriesIndex,
        type = p.type || p.searchKey || accessor.chart.chartTypes[seriesindex].name;
    return type;
}

/**
 *
 * @param points
 * @param accessor
 * @return {string}
 */
function getPointRankType(points, accessor) {
    if (isDependentPointHighlightType(accessor.chart)) {
        return 'POINT';
    } else {
        return 'SEARCH_MODE';
    }
}
