// @zohocharts/zc/src/events/dataSelection/search-mode.js
/*
       List of search tool model
    */
import { d3_set } from '../../core/d3-constants';

export var QUADTREE_TYPES = d3_set([
        'line',
        'scatter',
        'bubble',
        'bubblepie',
        'threshold',
        'threshold-marker',
        'segment',
        'treemap',
        'wordcloud'
    ]),
    POLY_TYPES = d3_set(['area', 'arearange', 'web', 'sankeyLinkLinear', 'sankeyNodes']),
    BISECT_TYPES = d3_set([
        'bar',
        'bullet',
        'gantt',
        'heatmap',
        'waterfall',
        'boxplot',
        'barrange',
        'bar_segment',
        'barrange_segment',
        'boxplot_segment',
        'bar_triggeroffset'
    ]);

export var NONSERIES_TYPES = d3_set(['threshold']);

export var QUADTREE_NORM_TYPES = d3_set(['line', 'scatter', 'threshold']),
    QUADTREE_SIZE_TYPES = d3_set(['bubble']),
    QUADTREE_PIE_TYPES = d3_set(['bubblepie']),
    QUADTREE_BOUND_TYPES = d3_set(['treemap']),
    QUADTREE_TEXT_TYPES = d3_set(['wordcloud']);

export var BISECT_NORM_TYPES = d3_set(['bar', 'barrange', 'bullet', 'gantt', 'heatmap', 'waterfall']),
    BISECT_LEVEL_TYPES = d3_set(['boxplot']),
    BISECT_SEGMENT_TYPES = d3_set(['bar_segment', 'barrange_segment', 'boxplot_segment']),
    BISECT_QUADTREE_TYPES = d3_set(['bar_triggeroffset']);

export var SEGMENT_TYPES = d3_set(['segment', 'whisker']);

export var PATH_TYPES = d3_set(['sankeyLinkBezier', 'polar_bar']);
export var FLOW_TYPES = d3_set(['sankey']);

/**
 *
 * @param type
 * @param sharedCriteria
 * @return {"bisect" | "PATH" | "POLY" | "QUADTREE" | "BISECT"}
 */
export function getSearchMode(type, sharedCriteria) {
    if (sharedCriteria && !NONSERIES_TYPES.has(type)) {
        return 'bisect';
    } else if (PATH_TYPES.has(type)) {
        return 'PATH';
    } else if (POLY_TYPES.has(type)) {
        return 'POLY';
    } else if (QUADTREE_TYPES.has(type)) {
        return 'QUADTREE';
    } else if (BISECT_QUADTREE_TYPES.has(type)) {
        return 'BISECT_QUADTREE';
    } else if (BISECT_TYPES.has(type)) {
        return 'BISECT';
    } else if (type === 'zc-marker') {
        return 'QUADTREE';
    }
}

export function getFineSearchKey(mode) {
    if (mode == 'QUADTREE') {
        return 'bubble';
    } else {
        return 'bar';
    }
}

/**
 *
 * @param type
 * @return {string}
 */
export function getChartModel(type) {
    // MODELS ['FLOW', 'COLUMN', 'HIERARCHY', "CONTINUOUS"]

    if (FLOW_TYPES.has(type)) {
        return 'FLOW';
    }
}
