// @zohocharts/zc/src/events/dataSelection/selection-manager.js
import { SELECTED_NEW, SELECTED_OLD, SELECTED_MERGED, SELECTED_NULL } from './selection-types';

export class SelectionManager {
    constructor(chart, eventM) {
        this.chart = chart;
        this.eventManager = eventM;
        this.nearest = null;
        this.selectedEvent = null;
        this.selectionSet = [];
        this.newSelectionSet = [];
        this.oldSelectionSet = [];
        this.mergedSelectionSet = [];
        this.rejectionSet = [];
        this.oldSelectionSet = [];
        this.oldRejectionSet = [];
    }
    add(selections) {
        if (this.selectionSet.length) {
            this.mergedSelectionSet = this.intersection(selections, this.selectionSet, SELECTED_MERGED);
            this.oldSelectionSet = this.difference(this.selectionSet, this.mergedSelectionSet, SELECTED_OLD);
            this.newSelectionSet = this.difference(selections, this.mergedSelectionSet, SELECTED_NEW);
        } else if (selections.length) {
            this.newSelectionSet = selections;
            this.updateID(this.newSelectionSet, SELECTED_NEW);
        }
        this.selectionSet = selections;
        this.selectedEvent = this.eventManager.eventType;
    }
    remove() {
        this.updateID(this.mergedSelectionSet, SELECTED_NULL);
        this.mergedSelectionSet = [];
        this.oldSelectionSet = [...this.selectionSet];
        this.selectionSet = [];
    }
    updateID(set, id) {
        set.forEach(d => {
            d.selected = id;
        });
    }
    difference(set1, set2, id) {
        let difference = [];
        if (set2.length) {
            set1.forEach(d0 => {
                set2.forEach(d1 => {
                    if (d0._key == d1._key && !(d1.selected == SELECTED_MERGED)) {
                        difference.push(d0);
                    }
                });
            });
        } else {
            difference = set1;
        }

        this.updateID(difference, id);
        return difference;
    }
    intersection(set1, set2, id) {
        if (this.eventManager.eventType != this.selectedEvent) {
            return [];
        }
        let intersection = [];
        set1.forEach(d0 => {
            set2.forEach(d1 => {
                if (d0._key && d0._key == d1._key && (d1.selected == SELECTED_NEW || d1.selected == SELECTED_MERGED)) {
                    intersection.push(d0);
                }
            });
        });
        this.updateID(intersection, id);
        return intersection;
    }
}
