// @zohocharts/zc/src/axis/tickFilter/node-selection.js
import { d3_merge, d3_map } from '../../core/d3-constants';
import { ELEMENT_SELECTOR, NON_PLOT_MODULES } from './node-selection-mapping';

export function selectTickFilterNodes(chart, parent, tag, query, chartTypes, axisType) {
    var result = {
        selected: d3_map(),
        unSelected: d3_map(),
        reset: d3_map()
    };
    var preSelection = parent;

    var modules = axisType
        ? d3_merge([['tick', 'label', 'boundbox', 'connector', 'bar.levelMarker'], chartTypes])
        : d3_merge([['label', 'boundbox', 'connector'], chartTypes]);

    modules.map(d => {
        var parentWrap = (parent && parent.get(d)) || getParentWrap(d, chart, axisType),
            element = ELEMENT_SELECTOR.get(d),
            filterChildElement = ELEMENT_SELECTOR.get(`${d}_child`),
            attr = tag == 'row' ? 'levelbyrow' : 'levelbycolumn';
        preSelection = parent && !NON_PLOT_MODULES.has(d);
        var filter = selectAllByQuery(parentWrap, attr, query, element, preSelection, filterChildElement);

        result.selected.set(d, filter.selected);
        result.unSelected.set(d, filter.unSelected);

        if (d == 'tick') {
            result.reset.set(d, filter.reset);
        }
    });

    return result;
}

function getParentWrap(module, chart, axisType) {
    if (module == 'label' || module == 'boundbox' || module == 'connector') {
        return chart.staticPaneSVG.select('g.labelholder');
    } else if (module == 'tick' || module == 'axislabel') {
        return chart.wrap.selectAll(`.${axisType}-axis`);
    } else {
        return chart.svg.select('.seriesholder');
    }
}
export function selectAllByQuery(parent, attr, value, element, preSelection, filterChildElement) {
    var querySelector = preSelection ? 'filter' : 'selectAll';

    var pointsSelected = parent[querySelector](`${element}[${attr}^="${value}"]`),
        pointsNotSelected = parent[querySelector](`${element}:not([${attr}^="${value}"])`),
        reset = parent[querySelector](`${element}`);

    if (filterChildElement) {
        (pointsSelected = pointsSelected[querySelector](`${filterChildElement}`)),
            (pointsNotSelected = pointsNotSelected[querySelector](`${filterChildElement}`));
    }

    return {
        selected: pointsSelected,
        unSelected: pointsNotSelected,
        reset: reset
    };
}
