// @zohocharts/zc/src/axis/tickFilter/node-selection-mapping.js
import { d3_map, d3_set } from '../../core/d3-constants';
import { charttypenames } from '../../core/main-zc-init';

export var ELEMENT_SELECTOR = d3_map();

export var STYLE_SELECTOR = d3_map();

export var ATTR_MAP = d3_map();

export var PLOT_MODULES = d3_set();
export var NON_PLOT_MODULES = d3_set();

export var SERIES_TYPES = d3_set();

ATTR_MAP.set('row', 'levelByRow');
ATTR_MAP.set('column', 'levelByColumn');

ELEMENT_SELECTOR.set('bar', 'path.zc_bar');
ELEMENT_SELECTOR.set('bar.levelMarker', 'path.zc_levelMarker');
ELEMENT_SELECTOR.set('area', 'path.area');
ELEMENT_SELECTOR.set('line', 'path.line');
ELEMENT_SELECTOR.set('scatter', 'g.scatter');
ELEMENT_SELECTOR.set('bubble', 'g.scatter');
ELEMENT_SELECTOR.set('pie', 'path.zc_pie');
ELEMENT_SELECTOR.set('funnel', 'path.zc_funnel');
ELEMENT_SELECTOR.set('pyramid', 'g.pyramidsection');
ELEMENT_SELECTOR.set('sunburst', 'path.zc_sunburst');
ELEMENT_SELECTOR.set('treemap', 'path.zc_treemap');
ELEMENT_SELECTOR.set('packedbubble', 'circle.zc_packedbubble');
ELEMENT_SELECTOR.set('wordcloud', 'text.zc_wordcloud');
ELEMENT_SELECTOR.set('web', 'path.area');
ELEMENT_SELECTOR.set('tick', 'g.tick');
ELEMENT_SELECTOR.set('tick_child', 'text.ticklabel');
ELEMENT_SELECTOR.set('axislabel', 'text.axislabel');
ELEMENT_SELECTOR.set('label', 'text');
ELEMENT_SELECTOR.set('boundbox', 'path.zc_label_background');
ELEMENT_SELECTOR.set('connector', 'path.zc_label_connector');

STYLE_SELECTOR.set('area', ['opacity', 'fillOpacity']);

PLOT_MODULES.add(['label', 'series', 'boundbox', 'connector']);
NON_PLOT_MODULES.add(['tick']);

export function getModuleType(key) {
    if (charttypenames.has(key)) {
        return 'series';
    } else {
        return key;
    }
}
