// @zohocharts/zc/src/core/browser-detect.js
//browser-detect.js
import { Registry } from './zc-registry';

export var $Browser = new (function() {
    this.NETSCAPE = 0;
    this.FIREFOX = 0;
    this.OPERA = 0;
    this.IE = 0;
    this.SAFARI = 0;
    this.MOZILA = 0;
    this.CHROME = 0;
    this.NONIE = 1;
    this.WEBKIT = 0;
    var agent = navigator.userAgent.toLowerCase();
    this.OSX = agent.indexOf('mac') != -1;
    this.WINDOWS = agent.indexOf('windows') != -1;
    this.LINUX = agent.indexOf('linux') != -1;
    this.IPHONE = agent.indexOf('iphone') != -1;
    this.IPAD = agent.indexOf('ipad') != -1;
    this.ANDROID = agent.indexOf('android') != -1;
    this.IS_IOS = this.IPHONE || this.IPAD;
    this.isTouchDevice = 'ontouchstart' in window || navigator.maxTouchPoints > 0 || navigator.msMaxTouchPoints > 0;
    this.isMouseDevice = !(this.IS_IOS || this.ANDROID || agent.indexOf('mobile') != -1);
    this.uiwebview = this.IS_IOS && !/safari/.test(agent);
    this.PHANTOMJS = agent.indexOf('phantomjs') !== -1;
    this.WKWebView = this.uiwebview && agent.indexOf('wkwebview') != -1;
    var prefix = '';
    this.detect = function() {
        var agent = navigator.userAgent;
        var uaLen = agent.length;
        var vendor = navigator.vendor;
        var operaObject = window.opera;
        var versionSub = '';
        if (operaObject) {
            this.OPERA = 1;
            prefix = 'o';
        } else if (agent.indexOf('Chrome') > -1) {
            this.CHROME = 1;
            versionSub = 'Chrome/';
            prefix = 'webkit';
        } else if (agent.indexOf('Firefox') > -1) {
            this.FIREFOX = 1;
            versionSub = 'Firefox/';
            prefix = 'moz';
        } else if (agent.indexOf('MSIE ') > -1 || agent.indexOf('Trident/') > -1) {
            this.IE = 1;
            versionSub = 'MSIE ';
            prefix = 'ms';
        } else if (vendor.indexOf('Apple') > -1) {
            this.SAFARI = 1;
            versionSub = 'Version/';
            prefix = 'webkit';
        } else {
            versionSub = 'Unknown Browser';
        }
        this.VERSION = agent.substring(agent.indexOf(versionSub), uaLen).match(/[0-9\.]+/)[0];
    };
    this.detect();
    this.NONIE = !this.IE;
    this.WEBKIT = this.SAFARI || this.CHROME;
    this.prefix = {
        lowercase: prefix,
        css: prefix ? `-${prefix}-` : prefix, //prefix = "" if browser is unknown
        js: prefix ? prefix[0].toUpperCase() + prefix.substr(1) : prefix //prefix = "" if browser is unknown
    };
})();

Registry.setComponent('browser', $Browser);
