// @zohocharts/zc/src/core/charttype-helpers.js
//charttype-helpers.js

import { Registry } from '../core/zc-registry';
import { charttype, chartcategory } from './main-zc-init';
import { arrayIncludes, getPropVal, objectValues } from '../utils/zcutils';
import { isFacet } from '../axis/hierarchy-axis-utils';
import { d3_set } from './d3-constants';
import { isMultiColoring } from '../utils/colorutils';

export function isAreaChartType(type) {
    // return [$ZC.charttype.area, $ZC.charttype.stkdarea, $ZC.charttype.arearange].indexOf(type) > -1;
    return type === charttype.area || type === charttype.stkdarea || type === charttype.arearange;
}

/**
 *
 * @param {Number} type chart type
 * @return {boolean}
 */

export function isBubbleChartType(type) {
    return chartcategory.bubble.has(type) || chartcategory.bubblePie.has(type);
}

export function getChartTypeByIndex(chart, paneid, index, series) {
    if (chart.dataObject.categoryHierarchy && chart.dataObject.hierarchical.pivot) {
        if (!paneid) {
            if (series) {
                paneid = `${series.row}_${series.column}`;
            } else {
                paneid = '0_0';
            }
        }
        let gridOptions = chart.dataObject.hierarchical.gridOptions;
        return gridOptions[paneid].chartTypes[index] || gridOptions[paneid].chartTypes[0];
    } else {
        return chart.chartTypes[index] || chart.chartTypes[0];
    }
}

export function getUniqueChartTypes(chart) {
    return d3_set(chart.distinctChartTypes);
}

export function isCategoryOf(chartObj, chartType, categoryType) {
    if (isFacet(chartObj)) {
        let categoryKeys = { axis: 'axis', polarAxis: 'polarAxis', nonAxisMultiSeries: 'nonaxismultiseries', hierarchy: 'hierarchy' };
        return arrayIncludes(chartcategory[categoryKeys[categoryType]], chartType);
    }

    let savedKeys = {
        axis: 'isAxisCategory',
        polarAxis: 'isPolarAxisCategory',
        nonAxisMultiSeries: 'isNonAxisMultiSeriesCategory',
        hierarchy: 'isHierarchy'
    };
    return chartObj.dataObject[savedKeys[categoryType]];
}

export function isMultiColorChart(chartName, chartObj) {
    var multiColoring =
            /(bar|scatter|bubble$|bullet|gantt|wordcloud|waterfall|boxplot|barrange)/.test(chartName) &&
            isMultiColoring(chartObj.systemConf.chart.plot.plotoptions[chartName], chartObj),
        defaultMultiColoring = /(^pie|funnel|pyramid|sankey|dial)/.test(chartName);
    return multiColoring || defaultMultiColoring;
}

export function isAlluvialChartType(chartObj) {
    return chartObj.distinctChartTypes[0] == 'sankey' && getPropVal(chartObj.userdata, 'chart.plot.plotoptions.sankey.hierarchical');
}

export function isPolarSupportedChart(chartObj, type) {
    return chartObj.dataObject.isPolarAxisCategory && chartcategory.polarSupported.includes(charttype[type]);
}

Registry.setComponent('utils.isCategoryOf', isCategoryOf);
Registry.setComponent('utils.getChartTypeByIndex', getChartTypeByIndex);
