// @zohocharts/zc/src/core/chart-instance-manager.js
//$Id$
/*
 * Main method to generate a chart or charts
 * containers,datas param can be an array if there have to be multiple chart generation.
 */
import { isArray, isElementOnDocument } from './../utils/zcutils.js';
import { d3_select } from './d3-constants';

export var charts = function(containers, datas, dimension) {
    var isMulti = isArray(containers) && isArray(datas),
        isOverview = datas.__$overview;
    if (isMulti) {
        return getMultiCharts(containers, datas);
    } else {
        var chartObj = getChartObject(containers, datas);
        if (isOverview) {
            adjustOverviewDim(chartObj, dimension);
        }
        if (chartObj) {
            chartObj.isMiniature ? chartObj.sparkline.draw() : chartObj.draw();
        }
        return chartObj;
    }
};

export var maps = function(containers, data) {
    if (!containers.isD3Selection) {
        containers = d3_select(containers);
    }

    return getMapObject(containers, data);
};

function getMapObject(containers, data) {
    var mapObj = new maps.GeoMap(containers, data);
    return mapObj;
}
/*
 * This is the actual chart object init method
 * here the data is binded with respective selected container
 * Note: Only the chart object is initialized here, but the chart generation or rendering will not happen.
 * */
function getChartObject(container, data) {
    if (!container.isD3Selection) {
        container = d3_select(container);
    }

    if (!isElementOnDocument(container.node())) {
        // if (!document.body.contains(container.node())) {
        this.console.warn('No container');
        return null;
    }

    var dataObj = container.datum(data);
    var chart = null;

    /*
     * Each chart instance is instantiated with chartRendered flag false
     * chart.initialize is defined in chart-constructor.js
     * */
    dataObj.each(function(data) {
        var chartHolder = d3_select(this);
        chart = charts.get(chartHolder, data); // get chartobj
        chart.showLoading(chartHolder);
    });
    return chart;
}
/*
 * when it is multi charts instantiation, the chart rendering is queued here.
 */
function getMultiCharts(containers, datas) {
    if (!(isArray(containers) && isArray(datas))) {
        return;
    }

    var chartObjs = containers.map(function(container, i) {
        return getChartObject(container, datas[i]);
    });

    var executedChartObjs = [];
    //unexecutedChartObjs are reversed just for the convenience of using Array.pop() method to remove first rendered chart object.
    var clonedObjs = chartObjs.slice();
    var unexecutedChartObjs = clonedObjs.reverse();
    drawEfficiently(unexecutedChartObjs, executedChartObjs, chartObjs);
    return chartObjs;
}

function drawEfficiently(unexecutedChartObjs, executedChartObjs, chartObjs) {
    var len = unexecutedChartObjs.length,
        chartObj = unexecutedChartObjs[len - 1];

    // all charts are rendered
    if (!len) {
        chartObjs.chartRendered = true;
    }

    if (len > 0 && chartObj.drawStarted != true && isElementOnDocument(chartObj.container.node())) {
        chartObj.draw();
        chartObj.drawStarted = true;
        checkStatusNCallNext(unexecutedChartObjs, executedChartObjs, chartObjs);
    }
}

/*
 * Executed/rendered chart objects are popped and pushed to the executed object list.
 * Waits till next chart object is rendered.
 * */
function checkStatusNCallNext(unexecutedChartObjs, executedChartObjs, chartObjs, timer) {
    var len = unexecutedChartObjs.length,
        chartObj = unexecutedChartObjs[len - 1];

    if (chartObj.chartRendered || chartObj.breakOnErr) {
        executedChartObjs.push(unexecutedChartObjs.pop());
        drawEfficiently(unexecutedChartObjs, executedChartObjs, chartObjs);
    } else {
        if (timer != null) {
            clearTimeout(timer);
        }

        var timer = setTimeout(function() {
            checkStatusNCallNext(unexecutedChartObjs, executedChartObjs, chartObjs, timer);
        }, 10);
    }
}

function adjustOverviewDim(chartObj, _dimension) {
    var _canvas = chartObj.canvasarea,
        _chart = chartObj.chartarea,
        _title = chartObj.titlearea;

    _canvas.width = _dimension.width;
    _chart.width = _dimension.width;
    _title.width = _dimension.width;
    _canvas.height = _dimension.height;
    _chart.height = _dimension.height;
    _title.height = _dimension.height;
}
