// @zohocharts/zc/src/core/chart-constructor.js
//$Id$
//chart-constructor.js
//Returns the chart object after setting all the required properties and methods

// (function(exports) {

// "use strict";
import {
    arrayIncludes,
    defined,
    getPropVal,
    inherit,
    isArray,
    isObjectPropDefined,
    mergeJSON,
    objectValues,
    pick,
    setPropVal
} from './../utils/zcutils.js';
import { charts } from './chart-instance-manager';
import { VizLayout } from './main-constructor';
import { axis, canvas_plot_renderer, chartcategory, plot, renderer_helpers, scale, stack, tooltip } from './main-zc-init';
import { DataProcessor } from './data-processor';
import { setDefaultOptions } from './defaulttheme';
import { stateChangeListener } from '../events/eventActions';
import { dataparser } from './data-parser';
import { PreProcessor } from './data-processor/pre-data-processor';
// import { tooltip_block_helpers } from '../tooltip/tooltip-blockPosition';
import { colorscale_helpers } from '../axis/colorscale-helpers';
import { getAxisRotated } from '../axis/axisutils';
import { Registry } from './zc-registry';
import { seriesIs } from '../events/events-utils';
import { DIAL_CHARTNAME, PIE_CHARTNAME, FUNNEL_CHARTNAME, PYRAMID_CHARTNAME } from '../core/chartname-constants';
import { ORDINAL_DATATYPE } from './datatype-constants';
import { addProcessorPolyfills } from './data-processor/data-processor-polyfill';
import { d3_range } from '../core/d3-constants';
import { createElement } from '../utils/zcsvgutils';
// import { initTooltip, updateTooltip } from '../events/visualEffects/controllers/tooltip_c';
import { addClipToTracker } from '../area/webClippath';
import { BehaviouralEvents } from '../behaviours/behavioural-events';
import { _ZoomToDomain } from '../behaviours/visual-effects/chart/controllers/zoom-to-domain';
import { _BrushToDomain } from '../behaviours/visual-effects/chart/controllers/brush-to-domain';
import { isFacet } from '../axis/hierarchy-axis-utils';
import { getUniqueChartTypes, isAlluvialChartType } from './charttype-helpers';
import { mathMax } from './zc-constants';
import { applyEffect } from '../effects/zceffects';
import { updateScaleForPivot } from '../axis-scale/pivot-scale-update-helpers';
// import { removeDummyData } from './data-processor/stack-helpers';
import { addSeries } from '../plot/addRemoveSeries/addSeries.js';
import { removeSeries } from '../plot/addRemoveSeries/removeSeries.js';
import { renderDom } from '../mark/dom-render.js';
import { invokeFunction } from './../utils/zcutils.js';
import { createScaleComponetForParser } from '../parser/parser-add-scale.js';
import { getCustomPluggedChartTypes } from './series-parser-helpers.js';
import { ZC_PLUGGABLE_VIZ } from './main-zc-init';
// export * from './../scroll/defaultScroll';
// TODO renaming exports or $ZC
charts.get = function(chartHolder, data) {
    return new Chart(chartHolder, data);
};

export function Chart(container, data) {
    let self = this;
    this.container = container;
    this.userdata = data;
    this.dataObject = null;
    this.dataset = null;
    this.svg = null;
    this.seriesColor = [];
    this.axes = {
        x: {
            levels: [],
            branches: [],
            clusters: [],
            leaves: []
        },
        y: {
            groups: [],
            branches: [],
            clusters: [],
            leaves: []
        }
    };
    // this.renderer = [];
    this.ismultiy = null;
    this.plotarea_approx = null;
    this.nochartdata = false;
    this.chartRendered = false;
    this.plot = {};
    this.events = {};

    this.viewPortExtent = null;
    this.isMiniature = this.isMiniature || getPropVal(data, 'chart.plot.miniature') ;
    this.customVisual = data._viewParser && data.renderOldConfig == false;
    // this.legend = {
    //     component: null
    // };

    // extend VizLayout
    if (this.isMiniature) {
        canvas_plot_renderer.sparkline(this);
    } else {
        VizLayout.call(this, container, data);
        this.initialize();
    }
    
}

inherit(Chart, VizLayout);
Chart.helpers = VizLayout.helpers;

Chart.prototype.initialize = function() {
    var chartObj = this;

    // CURRENT_CHART_COUNT++;
    // chart.id = "ZC_" + CURRENT_CHART_COUNT;
    // chart.idNumeric = CURRENT_CHART_COUNT;

    try {
        // var data = this.userdata,
        // dataProcessorHelpers = chart.dataProcessorHelpers = DataProcessor.helpers;

        // chart.createHelperElements();
        // chart.handleHiddenContainer();
        // chart.chartFontProp = {};
        // chart.chartWidth = chart.getChartDimension("width");
        // chart.chartHeight = chart.getChartDimension("height");
        // Chart.helpers.initializeChartHolders(chart);
        // chart.setCanvasDimension();

        // get all chart types
        // chart.chartTypes = dataProcessorHelpers.getAllChartTypes(data, "object");
        // chart.globalProp = setGlobalPropByUserdata(data, {
        //     "width": chart.canvasarea.width,
        //     "height": chart.canvasarea.height
        // }, chart.chartFontProp);

        // chart.themeoptions = chart.getThemeOptions(chart.globalProp);

        // (!data.metadata) ? setDefaultOptions("metadata", data) : null; // If metadata is unavailable, get default metadata
        chartObj.timerObj = new Array(0);
        chartObj.processUserData(chartObj.themeoptions);
        chartObj.setChartDimension();

        let legendHighlight = Registry.getComponent('LegendHighlight');
        chartObj.legendHighlight = new legendHighlight(chartObj, chartObj.svg);

        VizLayout.helpers.constructChartGroups(chartObj);
    } catch (err) {
        VizLayout.helpers.handleErrors(chartObj, err);
    }
};

/**
 * @returns {void}
 */

Chart.prototype.draw = function() {
    var chart = this,
        plotBgData = this.systemConf.chart.plot.background;
    if (plotBgData.image.lockAspectRatio && plotBgData.imageUrl) {
        let img = new Image();
        this.plot.hasBgAspectRatio = true;
        let plotBgObj = (this.plot.backgroundImg = {});
        img.onload = function() {
            plotBgObj.width = this.width;
            plotBgObj.height = this.height;
            chart.render(false);
        };
        img.src = plotBgData.imageUrl;
    } else {
        this.render(false);
    }
};

/**
 *
 * @param {boolean} update
 */

Chart.prototype.render = function(update) {
    var chart = this;
    try {
        chart.updateCanvasArea();

        var systemConfig = chart.systemConf;
        chart.applyChartProperties();
        if (chart.nochartdata) {
            chart.hideLoading(chart.container);
            chart.handleNoData(systemConfig.noDataHandler);
            chart.chartRendered = true;
            stateChangeListener(chart, systemConfig.canvas.events, 'chart', update ? 'onreload' : 'onload');
        } else {
            chart.clearNoData(systemConfig.noDataHandler);
            chart.getChartSeriesColor();
            chart.isAxisCategory = chart.dataObject.isAxisCategory;
            chart.isNonAxisMultiSeriesCategory = chart.dataObject.isNonAxisMultiSeriesCategory;
            chart.isPolarAxisCategory = chart.dataObject.isPolarAxisCategory;

            chart.parseScale();
            chart.zscale = chart.scales[0] && chart.scales[0].getScale();

            Registry.getComponent('legend.render')(chart, update);  //TODO: clr scale constructed in legend. Need to create standalone clr scale
            tooltip.block_helpers && tooltip.block_helpers.updatePlotarea4BlockTooltip(chart);
            prePlotRenderings(chart, systemConfig, update);

            if (chart.plot.renderer) {
                chart.plot.renderer.draw(update);
            } else {
                var renderer = (chart.plot.renderer = plot.render(chart, update));
                renderer.draw(false);
            }

            //Handling viewparser custom visual marks
            if(chart.viewFlow && chart.viewFlow.marks._custom.length){
                let timerObj;
                chart._customVisualRendered = false;
                chart.viewFlow.marks._custom.map((mark, i) => {
                    let parser = invokeFunction(mark.mark.parser, chart);
                    renderDom(parser, null, chart.viewFlow);
                })

                chart.CustomVisualPromise = new Promise(resolve => {
                    chart.timerObj[chart.timerObj.length] = timerObj = setInterval(function(){
                        if((chart.viewFlow.transition.start && chart.viewFlow.transition.end) || (!chart.viewFlow.transition.start)){
                            clearInterval(timerObj);
                            resolve();
                        }
                    }, 100)
                    
                })
                
            }

            // fix the legend position based on plot area
            // chart.fixLegendPosition();

            
            const isSmoothScrollEnabled = Registry.getComponent('scroll.isSmoothScrollEnabled');
            if (isSmoothScrollEnabled(chart)) {
                chart.divScroll.init(chart);
            }
            if (chart.isAxisCategory) {
                chart.drawThreshold(update);
            }
            chart.handleEvent(update);
            chart.drawPlotareaClippath();

            chart.positionEventTrackerElement(chart.wrap.selectAll('rect.dummy-rect-filter'));

            // stateChangeListener(chart, systemConfig.canvas.events, "chart", (update)?"onreload":"onload");
        }
    } catch (err) {
        VizLayout.helpers.handleErrors(chart, err);
    }
};

function prePlotRenderings(chart, systemConfig, update) {
    if (isObjectPropDefined(chart.systemConf, 'notes') && chart.systemConf.notes.enabled) {
        //Adding block level annotations
        let annotation = Registry.getComponent('annotation');
        if (!annotation._empty) {
            annotation.addBlockAnnotations(chart);
        }
    }

    if (chart._isAxisCategory() || chart.dataObject.isPolarAxisCategory || chart.dataObject.hasLevelMarker) {
        chart.axes.rotated = getAxisRotated(systemConfig);
        if (chart.axes.rotated) {
            chart.axes.x.orient = 'left';
            chart.axes.y.orient = 'bottom';
        } else {
            chart.axes.x.orient = 'bottom';
            chart.axes.y.orient = 'left';
        }
        chart.ismultiy = chart.getAxesCount('y') > 1;
        chart.ismultix = chart.getAxesCount('x') > 1;
    }

    //brushing
    if (isObjectPropDefined(chart.systemConf.chart, 'scroll') && chart.systemConf.chart.scroll.enabled) {
        if (chart.scroll) {
            chart.scroll.update();
        } else {
            let ChartBrushScroll = Registry.getComponent('ChartBrushScroll');
            setDefaultOptions('scroll', chart.systemConf.chart, chart.globalProp);
            var scrollOrient = chart.systemConf.chart.scroll.orient,
                scrollType = !chart.axes.rotated ? 'y' : 'x';
            chart.scroll = new ChartBrushScroll(chart, 'x');
        }
    } else if (chart.scroll) {
        chart.scroll.destroy();
        chart.scroll = null;
        chart.behaviouralEvents.eventManager.behaviours.CHART_SCROLL = null;
        // #ZC2320
        delete chart.behaviouralEvents.eventManager.behaviours.CHART_SCROLL;
        delete chart.scroll;
    }
    //aspect ratio
    if (chart.plot.backgroundImg) {
        let imgWidth = chart.plot.backgroundImg.width,
            imgHeight = chart.plot.backgroundImg.height,
            aspectRatio = imgWidth / imgHeight,
            widthT = mathMax(chart.chartWidth, chart.chartHeight) * aspectRatio,
            heightT = widthT / aspectRatio;

        chart.systemConf.chart.width = widthT;
        chart.systemConf.chart.height = heightT;
    }

    const isSmoothScrollEnabled = Registry.getComponent('scroll.isSmoothScrollEnabled');
    if (isSmoothScrollEnabled(chart)) {
        let smothScroll = Registry.getComponent('smoothScroll');
        var div_scroll = new smothScroll();
        chart.divScroll = div_scroll;
        div_scroll.initialize(chart);
    }

    // Axis renderer
    stateChangeListener(chart, systemConfig.chart.axes.events, null, 'init');

    //plot border ZC1384
    chart.drawPlotBorder();



    if (chart._isAxisCategory() ||(chart.customVisual && chart.userdata.axes && chart.userdata.axes.length)) {
        axis.render(chart, chart.chartWidth, chart.chartHeight, update);
    }

    //Create scale component that was not created in axis.render
    if(chart.viewFlow){
        chart.viewFlow.bound = {
            canvasarea: chart.canvasarea, 
            chartarea: chart.chartarea, 
            plotarea:chart.plotarea
        }
    }
    if(chart.viewFlow && chart.viewFlow.scales){
        for(let id in chart.viewFlow.scales){
            if(!chart.viewFlow.scales[id].component){
                chart.viewFlow.scales[id].component = createScaleComponetForParser(chart.viewFlow.scales[id].conf, chart);
                // let _parserConf = chart.viewFlow.scales[id].conf,
                // parser_expression_val = applyExpressionValueForParser(_parserConf, chart.viewFlow);

                // parser_expression_val.map(d => {
                //     _parserConf[d[0]] = d[1];
                // })

                // var newScaleComp1 = new Scale(null, 0, _parserConf, chart);
                // let newScaleObj = newScaleComp1.getScale();
                // chart.viewFlow.scales[id].component = newScaleObj;
            }
        }
    }

    // Polar axis rendering
    let hasPolarAxis = chart.dataObject.isPolarAxisCategory || seriesIs(chart.chartTypes[0].value, DIAL_CHARTNAME);
    if (hasPolarAxis) {
        axis.polar.render(chart);
    }

    if (chart.legend.isSingleLayout) {
        chart.zscale = chart.dataObject.hasZaxis && !chart.dataObject.isHierarchy ? chart.scales[0].create() : null;
    }
    stateChangeListener(chart, systemConfig.chart.axes.events, null, update ? 'onreload' : 'onload');

    chart.updatePlotareaProp();
    return new Promise((resolve, reject) => {
        // setTimeout(() => {
        //     resolve();
        // }, 0);
        resolve();
    });
}

/**
 *
 * @param baseSystemConf
 * @param update
 */

Chart.prototype.processUserData = function(baseSystemConf, update) {
    var chart = this,
        systemConfig,
        series,
        chartdata;

    chart.systemConf = systemConfig = mergeJSON(baseSystemConf, chart.userdata, true, true);
    series = getPropVal(systemConfig, 'seriesdata') || {};
    // chart.seriesdata = chartdata = simpleClone(series.chartdata);
    chart.seriesdata = chartdata = series.chartdata;

    let customPluggedTypes = getCustomPluggedChartTypes(chart, chart.userdata);

        if(customPluggedTypes.length){
            
            chart.customVisual = true;
            chart.pluginVisual = true;
            let pluginParser = ZC_PLUGGABLE_VIZ._types[customPluggedTypes[0]],
                config_ = pluginParser(chart.userdata.chart.plot.plotoptions[customPluggedTypes[0]] || {},chart.userdata);

            mergeJSON(chart.systemConf, config_, true, true);
        }

    // if (!update) {
    stateChangeListener(chart, systemConfig.canvas.events, 'chart', 'init');
    // }
    let _chartSpecGetter = Registry.getComponent('dataSplicer');
    if (chart.userdata.viewParser || chart.userdata._viewParser || chart.customVisual || chart.pluginVisual) {
        initViewParser(chart);
        _chartSpecGetter = Registry.getComponent('dataSplicer4ViewParser');
    }

    chart.dataSplicer = new _chartSpecGetter(chart.userdata, chart.systemConf, chart);

    chart.nochartdata = !defined(chartdata) || (defined(chartdata) && chartdata.length < 1);

    if (chart.nochartdata) {
        chart.seriesdata = null;
    } else {
        addColorColumnToMeta(chart); //ZC2635

        chart.dataset = new dataparser.get(systemConfig, chart);
        chart.preProcessor = new PreProcessor(chart);
        if (!chart.seriesdata.length) {
            chart.nochartdata = true;
            return;
        }
        chart.dataObject = chart.getDataObject();

        if(chart.viewFlow){
            chart.viewFlow.process(chart.seriesdata);
        }
        
        chart.nochartdata = chart.dataObject.totalnoofdatas < 1;
    }
};

/**
 *
 *
 * @returns {*}
 */

Chart.prototype.getDataObject = function() {
    var chart = this,
        chartType = chart.chartTypes[0].value,
        processorName,
        processor;

    let { dataObject, preProcessor } = this,
        { hierarchical } = dataObject;

    let isPivotType =
        (hierarchical && hierarchical.facetRendering) ||
        (preProcessor && preProcessor.hierarchical && preProcessor.hierarchical.facetRendering);
    (processorName =
        (!isPivotType && arrayIncludes(chartcategory.hierarchy, chartType) && !chart.isAxisedHierarchyChart) || 
        (chart.customVisual && chart.viewFlow.hierarchy)
            ? 'hierarchyDataProcessor'
            : arrayIncludes(chartcategory.axis, chartType) ||
              arrayIncludes(chartcategory.polarAxis, chartType) ||
              seriesIs(chartType, DIAL_CHARTNAME) ||
              arrayIncludes(chartcategory.flow, chartType) ||
              isPivotType ||
              chart.isAxisedHierarchyChart
            ? 'cartesianDataProcessor'
            : 'defaultDataProcessor'),
        (processor = Registry.getComponent(processorName));

    /*if (chart._isAxisCategory()) {
        processorName = 'cartesianDataProcessor';
    } else if (arrayIncludes(chartcategory.hierarchy, chartType)) {
        processorName = 'hierarchyDataProcessor';
    } else if (
        arrayIncludes(chartcategory.axis, chartType) ||
        arrayIncludes(chartcategory.polarAxis, chartType) ||
        seriesIs(chartType, DIAL_CHARTNAME) ||
        arrayIncludes(chartcategory.flow, chartType)
    ) {
        processorName = 'cartesianDataProcessor';
    } else {
        processorName = 'defaultDataProcessor';
    }

    processor = Registry.getComponent(processorName);*/

    addProcessorPolyfills(processor); //TODO: remove

    return new processor(chart.systemConf, chart.dataset, chart);
};

/**
 *
 * @param type
 * @returns {number}
 */
Chart.prototype.getAxesCount = function(type) {
    var chart = this,
        count = 1,
        dataSplicer = chart.dataSplicer,
        metadata_axes = dataSplicer.getEncode(),
        metadata_columns = dataSplicer.getColumns(),
        hierarchical = chart.dataObject.hierarchical;

    if (type === 'x') {
        var columnIndex = chart.dataset.getColumnIdx(type),
            originalDataType = metadata_columns[columnIndex].datatype,
            formatConfig = pick(metadata_columns[columnIndex][originalDataType], {}),
            formatFunction = formatConfig.subfunction,
            axisEnabled = chart.systemConf.chart.axes.xaxis.show;

        if (hierarchical) {
            let columnList = hierarchical.pivotLayout.getRootNodeList('column');
            count = columnList.getCount() + columnList.getLeavesCount();
        } else if (axisEnabled && isArray(formatConfig.groupby) && formatConfig.groupby.length) {
            count = formatConfig.groupby.length + 1;
        } else if (axisEnabled && formatConfig.groupby === 'AUTO') {
            count = DataProcessor.HierarchicalTimeDataAdapter.getAutoGroupOrder(formatFunction).length + 1;
        }
        chart.axes[type].count = count;
    } else {
        var metadataY = metadata_axes.y;

        if (hierarchical && hierarchical.pivot) {
            var rowList = hierarchical.pivotLayout.getRootNodeList('row'),
                totalNoOfRows = rowList.getCount() || 0,
                leavesCount = rowList.getLeavesCount();
            count = totalNoOfRows + leavesCount;
        } else {
            count = metadataY.length;
        }
        chart.axes[type].count = count;
    }
    return count;
};

/**
 *
 * @param update
 */

Chart.prototype.drawThreshold = function(update) {
    if (axis.threshold) {
        var chart = this;
        chart.cache.thresholdCustomPoints = []; // clear it
        axis.threshold(chart, update);
        applyEffect(chart.container.selectAll('g.threshold'), { chartObj: chart });
    }
};

//? Does this function needed in custom bundling without events
/**
 *
 * @param update
 */
Chart.prototype.handleEvent = function(update) {
    var chart = this,
        systemConfig = chart.systemConf,
        chartEvents = Registry.getComponent('ZC_ChartEvents');

    if (chartEvents._empty) { return };
    if (!update || !chart.eventHandler) {
        chart.createEventGroups(update);
        let trackerGroup = createElement('g', '.trackerGroup', [1], chart.eventWrap).attr('pointer-events', 'all'),
            trackerElement = createElement('rect', '.eventhandler', [1], trackerGroup);

        // TODO Hack
        trackerElement.node().__chart = chart;

        chart.eventHandler = new chartEvents(chart, trackerGroup);
        chart.eventHandler.trackerGroup = trackerGroup;
        chart.eventHandler.trackerElement = trackerElement;
        addClipToTracker(chart);
        chart.eventHandler.init();
        tooltip.initialize &&  tooltip.initialize(chart); //TODO code clean
    } else {
        chart.eventHandler.update();
        //update tooltip
        tooltip.update && tooltip.update(chart);
    }
    if (chart.eventHandler.isSingleSource()) {
        if (chart.renderer.length) {
            if (
                !(
                    chart.seriesAddRemoveDetails &&
                    chart.seriesAddRemoveDetails.seriesAddRemovePhase &&
                    chart.seriesAddRemoveDetails.seriesUpdatedAtEnd
                )
            ) {
                chart.dataObject.flattenData();
            }
        }
        chart.positionEventTrackerElement();
    }

    addClipToTracker(chart);

    const isSmoothScrollEnabled = Registry.getComponent('scroll.isSmoothScrollEnabled');
    if (isSmoothScrollEnabled(chart)) {
        chart.divScroll.defaultScroll(chart);
    }

    if (chart.behaviouralEvents) {
        chart.behaviouralEvents.update();
    } else if (systemConfig.__$overview !== true) {
        chart.behaviouralEvents = new BehaviouralEvents(chart);

        // TODO Hack
        let brushOverlay = chart.eventHandler.trackerGroup.selectAll('rect.overlay');
        if (!brushOverlay.empty()) {
            brushOverlay.node().__chart = chart;
        }
    }
    // render it before event
    chart.scroll && chart.scroll.render();

    chart.handleLegendEvents(); // Legend events
};

/**
 *
 * @param series
 * @param i
 * @param axisorder
 * @param updateAllAxes
 */

Chart.prototype.addRemoveSeries = function(series, i, axisorder, updateAllAxes) {
    var chart = this,
        systemConfig = chart.systemConf;

    chart.phase = 'legendFilter';
    chart.cache.correction = {};

    if (isFacet(chart)) {
        chart.cache.search = {};
        chart.plot.renderer.draw(true);
        return;
    }

    if (colorscale_helpers.isQuantileColorScale(chart)) {
        chart.phase = 'legendFilter-colorScale';
        chart.cache.search = {}; // ZC2390
        chart.plot.renderer.draw(true);
        return;
    }

    chart.dataObject.formattedData = {
        x: [],
        y: []
    };
    chart.dataObject.initialize('addorremoveseries');

    stateChangeListener(chart, systemConfig.chart.axes.events, null, 'init');
    var donotUpdateAxis = chart.racing && !chart.racing.keyFramedRace;
     if(!donotUpdateAxis){
    if (chart.isAxisCategory) {
        chart.updateZscale();
        var hierarchical = chart.dataObject.hierarchical;
        if (updateAllAxes) {
            chart.axes.y.groups.forEach(function(d, i) {
                // chart.updateScale([i, 0], 'y');
                chart.updateScale([d.scale ? i : 0, 0], 'y');
            });
            if (chart.dataObject.hierarchical) {
                var rowList = hierarchical.pivotLayout.getRootNodeList('row'),
                    rowCount = rowList.getCount();
                if (rowCount) {
                    scale.setRowScale(chart, 'row');
                }
            }
        } else {
            chart.updateScale(axisorder, 'y');
        }

        // #ZC2174 $[chart.viewPortExtent]
        chart.updateScale(0, 'x', chart.viewPortExtent, null, null, chart.racing ? true : false);
        if (hierarchical) {
            chart.axes.x.rangeScale = chart.axes.x.scale.copy();
            chart.axes.x.rangeScale.outerPaddingPx = chart.axes.x.scale.outerPaddingPx;
            if (chart.dataObject.hierarchical.groupByDiscreteTime) {
                chart.axes.x.rangeScale.domain(d3_range(chart.axes.x.scale.domain().length));
            }
            let rangeScaleUtils = hierarchical.setScaleForRange(chart.axes.x.rangeScale);
            hierarchical.pivotLayout.setRanges('column', rangeScaleUtils);
            chart.axes.x.levels.forEach(function(d, i) {
                chart.updateScale(axisorder, 'x', null, i + 1);
            });
            if (hierarchical.categoryTree) {
                axis.TickLabelFilter.updateTickEventHandler(chart, series.disabled, 'x');
            }
        }
        chart.drawThreshold(true);
        chart.updatePlotareaProp();
    } else if (chart.dataObject.isPolarAxisCategory) {
        axis.polar.render(chart, 'x');
    }
}
    

    stateChangeListener(chart, systemConfig.chart.axes.events, null, 'onreload');
    // update the events
    chart.events = {};
    chart.plot.renderer.draw(true);
    chart.drawPlotareaClippath();
    if (chart.isAxisCategory || chart.dataObject.isPolarAxisCategory) {
        chart.positionEventTrackerElement();
        if (chart.systemConf.chart.width || chart.systemConf.chart.height) {
            chart.divScroll.defaultScroll(chart, 'addorremoveseries');
        }
    }

    if (chart.dataObject.pushFlattenAnnotation && chart.eventHandler.isSingleSource()) {
        chart.dataObject.pushFlattenAnnotation();
    }

    // #ZC2374
    if (chart.behaviouralEvents) {
        chart.behaviouralEvents.update();
    }

    if (chart.scroll && !(chart.racing && chart.racing.isInitiated)) {
        chart.scroll.update();
    }
};
Chart.prototype.removeSeries = removeSeries;
Chart.prototype.addSeries = addSeries;
/**
 * Add data point to single series or multiple series
 * @param seriesidx
 * @param point
 * @param shift
 */
Chart.prototype.pushDataPoint = function(seriesidx, point, shift, paneIds) {
    var chart = this,
        isComboChart = chart.dataObject.isCombinationalChart,
        orderedSeriesdata = isComboChart ? renderer_helpers.arrangeSeriesRenderingOrder(chart) : chart.seriesdata,
        isPivot = chart.dataObject.pivot,
        hierarchical = chart.dataObject.hierarchical,
        facetRendering = hierarchical && hierarchical.facetRendering,
        gridlist,
        panes,
        containsStacked;

    if (facetRendering) {
        var pivotLayout = facetRendering && hierarchical.pivotLayout,
            layers = pivotLayout.getRootNodeList('layer');
        panes = layers.panes;
        gridlist = objectValues(panes);
    }

    // If series indices were not of ARRAY type, then convert it to array and process
    if (!Array.isArray(seriesidx)) {
        var seriesidx_arr = [],
            point_arr = [],
            rowidx_arr = [],
            colidx_arr = [];
        seriesidx_arr.push(seriesidx);
        point_arr.push(point);
        seriesidx = seriesidx_arr;
        point = point_arr;
        if (isPivot) {
            paneIds = [paneIds];
        }
    }
    //Reset cache for data
    chart.cache.data = {};
    chart.cache.search = {};
    // chart.cache.correction = {};
    seriesidx.forEach(function(d, i) {
        var paneid = isPivot ? `${paneIds[i][0]}_${paneIds[i][1]}` : '0_0',
            currOrdereddata = isPivot ? panes[paneid].data[d] : orderedSeriesdata[d],
            curSeriesidx = isComboChart && !isPivot ? currOrdereddata.seriesidx : d,
            series = DataProcessor.helpers.getSeriesdata(chart, paneid, curSeriesidx),
            renderer = isPivot ? hierarchical.gridOptions[paneid].renderer[curSeriesidx] : chart.renderer[curSeriesidx],
            stacked = renderer.commonRendererProp().stacked(),
            shiftPoint,
            shiftInfo,
            shiftCurSeries;

        shiftCurSeries = isArray(shift) ? shift[i] : shift;

        if (stacked) {
            containsStacked = true;
            series = stack.helpers.removeDummyData(series);
            series.data[0].push(point[i]);
        }

        series.data[0].push(point[i]);
        chart.dataObject.flattenData([
            {
                seriesindex: curSeriesidx,
                subseriesindex: 0,
                index: isPivot ? panes[paneid].index : 0,
                data: [[point[i]]],
                row: isPivot ? paneIds[i][0] : null,
                column: isPivot ? paneIds[i][1] : null,
                rendererIndex: renderer.commonRendererProp().rendererIndex()
            }
        ]);
        if (shiftCurSeries) {
            shiftPoint = series.data[0][0];
            var shiftedData = series.data[0].shift();
            shiftedData._filter = true;
            if (defined(renderer.shiftInfo?.())) {
                shiftInfo = renderer.shiftInfo();
                shiftInfo.count++;
            } else {
                shiftInfo = {};
                shiftInfo.count = 0;
            }
            shiftInfo.point = shiftPoint;
            shiftInfo.shiftedData = shiftedData;
            renderer.shiftInfo(shiftInfo);
        } else {
            renderer.shiftInfo(null);
        }

        // if (chart.ismultiy) {
        //     chart.updateScale(series, 'y');
        // }
    });
    if (isPivot) {
        //update pivot layout
        chart.dataObject.hierarchical.update(false);
        chart.dataObject.processedseriesData = chart.seriesdata = chart.dataObject.hierarchical.getSeries();
    }

    chart.dataObject.initialize();

    /*
    For stacked data, we need to remove the duplicate dummy data entries and need to add again the all data properly.
    Hence we are flatten all the series data again.
     */
    if (containsStacked) {
        chart.dataObject.flattenPoints = null;
        if (chart.renderer.length) {
            chart.dataObject.flattenData();
        }
    }
    //update x scale

    if (isPivot) {
        //update rows, columns scale
        updateScaleForPivot(chart);
        var axesCount = chart.axes.x.count;
        for (var i = 0; i < axesCount; i++) {
            chart.updateScale(null, 'x', null, i);
        }
        //update row scale
        //ZC2993 recalculate rowscale to find exact axis area on rotation of ticklabels
        updateScaleForPivot(chart, null, false);
    } else {
        chart.updateScale?.(null, 'x');
    }

    //update y scale
    if (isPivot) {
        var axesCount = chart.axes.y.count;
        for (var i = 0; i < axesCount; i++) {
            chart.updateScale(null, 'y', null, i);
        }
    } else {
        if (chart.ismultiy) {
            seriesidx.forEach(function(d, i) {
                var currOrdereddata = orderedSeriesdata[d],
                    curSeriesidx = isComboChart ? currOrdereddata.seriesidx : d,
                    series = chart.seriesdata[curSeriesidx];
                chart.updateScale?.(series, 'y');
            });
        } else {
            chart.updateScale?.(chart.seriesdata[0], 'y');
        }
    }

    //update plot and renderer
    chart.plot.renderer.draw(true);
    chart.drawThreshold(true);

    i;
    // chart.eventHandler.updateParams(chart);
};

/**
 *
 * @param seriesidx
 */

Chart.prototype.popDataPoint = function(seriesidx) {
    var chart = this,
        series = chart.seriesdata[seriesidx];

    series.data.forEach(function(d) {
        d.pop();
    });
    chart.dataObject.initialize();
    chart.updateScale?.(series, 'x');
    chart.updateScale?.(series, 'y');
    chart.plot.renderer.draw(true);
    chart.drawThreshold(true);
    // chart.eventHandler.updateParams(chart);
};

/**************** Zooom ********************/

/**
 *
 * @param seriesIndex
 * @param range
 */
Chart.prototype.zoomX = function(seriesIndex, range) {
    var chart = this;
    chart.__enableZoom('x');
    _ZoomToDomain(chart.behaviouralEvents.eventManager, range, 'x', seriesIndex);
};

/**
 *
 * @param seriesIndex
 * @param range
 */
Chart.prototype.zoomY = function(seriesIndex, range) {
    var chart = this;
    chart.__enableZoom('y');
    _ZoomToDomain(chart.behaviouralEvents.eventManager, range, 'y', seriesIndex);
};

/**
 *
 * @param seriesIndex
 * @param xrange
 * @param yrange
 */
Chart.prototype.zoomXY = function(seriesIndex, xrange, yrange) {
    var chart = this;

    chart.__enableZoom('xy');
    _ZoomToDomain(chart.behaviouralEvents.eventManager, [xrange, yrange], 'xy', seriesIndex);
};

Chart.prototype.zoomReset = function() {
    var chart = this;
    chart.behaviouralEvents.eventManager.behaviours.PLOT_ZOOM.instance.reset();
};

/**
 *
 * @param type
 */
Chart.prototype.brushON = function(type) {
    var chart = this;
    chart.__enableBrush(type);
};

/**
 *
 */
Chart.prototype.brushOFF = function() {
    var chart = this;
    chart.__enableBrush(null);
};

Chart.prototype.brushX = function(seriesIndex, domain) {
    var chart = this;
    chart.__enableBrush('x');
    _BrushToDomain(chart.behaviouralEvents.eventManager, domain, 'x', seriesIndex);
};

Chart.prototype.brushY = function(seriesIndex, domain) {
    var chart = this;
    chart.__enableBrush('y');
    _BrushToDomain(chart.behaviouralEvents.eventManager, domain, 'y', seriesIndex);
};

Chart.prototype.brushXY = function(seriesIndex, domainX, domainY) {
    var chart = this;
    chart.__enableBrush('xy');
    _BrushToDomain(chart.behaviouralEvents.eventManager, [domainX, domainY], 'xy', seriesIndex);
};

/**
 *
 * @param type
 */
Chart.prototype.changeZoomType = function(type) {
    var chart = this,
        systemConfig = chart.systemConf;

    if (!this.zoom) {
        //ZC1647 - M2.2 - To initialize systemconf.zoom only one time, else need to remove zoom obj from defaulttheme
        setDefaultOptions('zoom', systemConfig.chart, chart.globalProp);
    }
    systemConfig.chart.zoom.type = type;
};

function addColorColumnToMeta(chart) {
    let dataSplicer = chart.dataSplicer,
        systemConf = chart.systemConf,
        metadata_axes = dataSplicer.getEncode(),
        metadata_columns = dataSplicer.getColumns(),
        facetHierarchy = systemConf.facets,
        singleChartTypes = [PIE_CHARTNAME, FUNNEL_CHARTNAME, PYRAMID_CHARTNAME],
        hasSingleChartType = getUniqueChartTypes(chart)
            .values()
            .some(_ => arrayIncludes(singleChartTypes, _));

    const levelAxes = metadata_axes.level && metadata_axes.level[0].length ? metadata_axes.level[0] : metadata_axes.level;
    const isAlluvialChart = isAlluvialChartType(chart);
    const xidx = isAlluvialChart ? levelAxes.slice(-1)[0] : metadata_axes.x[0];
    var xColumn =  metadata_columns[xidx],
        xDataType = xColumn.datatype;

    if (
        facetHierarchy &&
        !(metadata_axes.clr && metadata_axes.clr.length) &&
        !(systemConf.facets.color && systemConf.facets.color.length) &&
        hasSingleChartType
    ) {
        systemConf.metadata.axes.clr = [metadata_axes.x[0]];
        setPropVal(chart.userdata, 'legend._dummy', true);
        setPropVal(chart.systemConf, 'legend._dummy', true);

        if (xDataType === ORDINAL_DATATYPE) {
            if (!isObjectPropDefined(xColumn.ordinal, 'nullformatting')) {
                setPropVal(xColumn, 'ordinal.nullformatting', 'null');
                setPropVal(xColumn, 'ordinal._dummyNullformatting', true);
            }
            let nullFormatter = isObjectPropDefined(xColumn.ordinal, 'nullformatting', true) || 'null';
            setPropVal(chart.userdata, 'legend.colorBand.nullAs', nullFormatter);
            setPropVal(chart.systemConf, 'legend.colorBand.nullAs', nullFormatter);
        }
    } else if (!hasSingleChartType && getPropVal(chart.userdata, 'legend._dummy')) {
        delete systemConf.metadata.axes.clr;
        setPropVal(chart.userdata, 'legend._dummy', undefined);
        setPropVal(chart.systemConf, 'legend._dummy', undefined);

        if (xDataType === ORDINAL_DATATYPE) {
            if (isObjectPropDefined(xColumn.ordinal, '_dummyNullformatting')) {
                setPropVal(xColumn, 'ordinal.nullformatting', undefined);
                setPropVal(xColumn, 'ordinal._dummyNullformatting', undefined);
            }

            setPropVal(chart.userdata, 'legend.colorBand.nullAs', undefined);
            setPropVal(chart.systemConf, 'legend.colorBand.nullAs', undefined);
        }
    }
}

function initViewParser(chart){
    let _viewParser = Registry.getComponent('viewParser');
    let bound = {
        canvasarea: chart.canvasarea, 
        chartarea: chart.chartarea, 
        plotarea:chart.plotarea
    }
    let viewParserConfig = chart.pluginVisual ? chart.systemConf : chart.userdata.viewParser || chart.userdata;
    chart.viewFlow = new _viewParser(viewParserConfig, chart.userdata, bound, chart);
    chart.parserConfig = chart.viewFlow.init();
}

export * from './data-processor';
export * from './data-processor/data-processor-helpers';

export * from './data-processor/pre-processor-helpers';
export * from './data-processor/convertTo3D';
export * from './data-processor/dataCount';
export * from './../plot/renderer';

/**
 *
 * @type {Chart}
 */

// })($ZC);
