// @zohocharts/zc/src/core/d3-constants.js
//d3-constants.js
// d3 selections
import * as d3 from '@zohocharts/d3';

export var d3_zc = d3, //export overall d3 object
    d3_select = d3.select,
    d3_selectAll = d3.selectAll,
    // de min-max

    d3_min = d3.min,
    d3_max = d3.max,
    d3_pairs = d3.pairs,
    d3_mean = d3.mean,
    /**
     * d3 axis axis with orients
     *
     * @type {Function}
     */
    d3_axis = (function() {
        var d3AxisOrients = {
            top: d3.axisTop,
            bottom: d3.axisBottom,
            left: d3.axisLeft,
            right: d3.axisRight
        };

        return function(orient, scale) {
            return d3AxisOrients[orient](scale);
        };
    })(),
    /**
     * mapping d3 easing methods
     *
     * @type {Function}
     */
    d3_ease = (function() {
        var easingTypes = {
            linear: d3.easeLinear,

            quad: d3.easeQuad,
            'quad-in-out': d3.easeQuadInOut,
            'quad-in': d3.easeQuadIn,
            'quad-out': d3.easeQuadOut,

            cubic: d3.easeCubic,
            'cubic-in-out': d3.easeCubicInOut,
            'cubic-in': d3.easeCubicIn,
            'cubic-out': d3.easeCubicOut,

            sin: d3.easeSin,
            'sin-in-out': d3.easeSinInOut,
            'sin-in': d3.easeSinIn,
            'sin-out': d3.easeSinOut,

            exp: d3.easeExp,
            'exp-in-out': d3.easeExpInOut,
            'exp-in': d3.easeExpIn,
            'exp-out': d3.easeExpOut,

            circle: d3.easeCircle,
            'circle-in-out': d3.easeCircleInOut,
            'circle-in': d3.easeCircleIn,
            'circle-out': d3.easeCircleOut,

            elastic: d3.easeElastic,
            'elastic-in-out': d3.easeElasticInOut,
            'elastic-in': d3.easeElasticIn,
            'elastic-out': d3.easeElasticOut,

            back: d3.easeBack,
            'back-in-out': d3.easeBackInOut,
            'back-in': d3.easeBackIn,
            'back-out': d3.easeBackOut,

            bounce: d3.easeBounce,
            'bounce-in-out': d3.easeBounceInOut,
            'bounce-in': d3.easeBounceIn,
            'bounce-out': d3.easeBounceOut
        };

        return function(type) {
            return typeof type === 'function' ? type : easingTypes[type] || d3.easeLinear;
        };
    })(),
    d3_ease_linear = d3.easeLinear,
    /**
     *
     * @param interpolator
     * @param tension
     * @returns {*}
     */
    d3_curves = (function() {
        var curves = {
            linear: d3.curveLinear,
            linearClosed: d3.curveLinearClosed,
            cardinal: d3.curveCardinal,
            cardinalClosed: d3.curveCardinalClosed,
            monotone: d3.curveMonotoneX,
            monotoneY: d3.curveMonotoneY,
            stepBefore: d3.curveStepBefore,
            stepAfter: d3.curveStepAfter,
            'linear-closed': d3.curveLinearClosed,
            'cardinal-closed': d3.curveCardinalClosed,
            'step-before': d3.curveStepBefore,
            'step-after': d3.curveStepAfter
        };

        // the above lines can be changed with below line.
        // curve = d3["curve" + interpolator[0].toUpperCase() + interpolator.slice(1)];

        return function(interpolator, rotated, tension) {
            var curve = curves[interpolator];

            if (interpolator == 'monotone' && rotated) {
                curve = curves.monotoneY;
            }

            if (tension !== undefined && tension !== null && curve.tension) {
                curve.tension(tension);
            }

            return curve;
        };
    })(),
    d3_tiles = (function() {
        var defaultMode = 'squarify';
        var tiles = {
            squarify: d3.treemapSquarify,
            slice: d3.treemapSlice,
            dice: d3.treemapDice,
            sliceDice: d3.treemapSliceDice,
            resquarify: d3.treemapResquarify
        };

        return function(mode) {
            return tiles[mode] || tiles[defaultMode];
        };
    })(),
    // d3 scales

    d3_scaleLinear = function() {
        var scale1 = d3.scaleLinear();
        scale1.ticks = function(count) {
            var d = this.domain();
            return d3.ticks(d[0], d[d.length - 1], count == null ? 10 : count);
        };
        return scale1;
    },
    d3_scaleOrdinal = d3.scaleBand,
    // /**
    //  * 
    //  * @param {Array} domain 
    //  * @param {Array} range 
    //  * @param {Function} hash - hash the key value of domain based on callback
    //  * @returns 
    //  */    
    // function ordinal(domain, range, hash) {
    //     //* overriding some scale functions;
    //     let scale = d3.scaleBand(),
    //         scale_domain = scale.domain,
    //         scale_range = scale.range;

    //     hash = hash || (_ => _);
    //     let scale$1 = function scale(_) {
    //         return scale(hash(_))
    //     }

    //     Object.assign(scale$1, scale);
        
    //     scale$1.hash = function (_) {
    //         return arguments.length ? (hash = _, scale$1) : hash;
    //     }
    //     scale$1.domain = function (_) {
    //         return arguments.length ? (scale_domain(hash ? _.map(d => hash(d)) : _), scale$1) : domain;
    //     }
    //     scale$1.range = function (_) {
    //         return arguments.length ? (scale_range(hash ? _.map(d => hash(d)) : _), scale$1) : domain;
    //     }

    //     return scale;
    // },
    d3_scaleTime = d3.scaleTime,
    d3_scalePoint = d3.scalePoint,
    d3_scaleOrdinalOrg = d3.scaleOrdinal, //change
    d3_scaleSqrt = d3.scaleSqrt,
    d3_scaleSymlog = d3.scaleSymlog,
    // bisector

    d3_bisector = d3.bisector,
    // sum

    d3_sum = d3.sum,
    // d3 colors

    d3_hsl = d3.hsl,
    d3_rgb = function() {
        var rgb = d3.rgb.apply(this, arguments);

        /**
         *
         * if the color is transparent, d3.rgb returns NaN, which throws error
         * in some browsers. eg: IE9
         *
         * So, if the color is NaN, we apply black color with 0 opacity.
         */
        if (isNaN(rgb.r) || isNaN(rgb.g) || isNaN(rgb.b)) {
            rgb = d3.rgb(0, 0, 0, rgb.opacity);
        }
        return rgb;
    },
    // d3 merge , map and ser

    d3_merge = d3.merge,
    d3_map = d3.map,
    d3_set = d3.set,
    // d3 behavior

    d3_zoom = d3.zoom,
    d3_zoomTransform = d3.zoomTransform,
    d3_zoomIdentity = d3.zoomIdentity,
    d3_drag = d3.drag,
    // d3 extent

    d3_extent = d3.extent,
    // event dispatch

    d3_dispatch = d3.dispatch,
    // d3 sort type

    d3_ascending = d3.ascending,
    d3_descending = d3.descending,
    // d3 shapes

    d3_line = d3.line,
    d3_lineRadial = d3.radialLine,
    d3_area = d3.area,
    d3_areaRadial = d3.radialArea,
    d3_arc = d3.arc,
    // layouts
    d3_hierarchy = d3.hierarchy,
    d3_pack = d3.pack,
    d3_pie = d3.pie,
    d3_partition = d3.partition,
    d3_treemap = d3.treemap,
    d3_tree = d3.tree,

    // d3 mouse
    // d3 interpolate

    d3_interpolate = d3.interpolate,
    d3_interpolateRgb = d3.interpolateRgb,
    d3_interpolateRound = d3.interpolateRound,
    // d3 range

    d3_range = d3.range,
    d3_tickStep = d3.tickStep,
    // d3 times

    // days
    d3_timeSunday = d3.timeSunday,
    d3_timeMonday = d3.timeMonday,
    d3_timeTuesday = d3.timeTuesday,
    d3_timeWednesday = d3.timeWednesday,
    d3_timeThursday = d3.timeThursday,
    d3_timeFriday = d3.timeFriday,
    d3_timeSaturday = d3.timeSaturday,
    // seconds, minute and hour
    d3_timeSeconds = d3.timeSeconds,
    d3_timeMinutes = d3.timeMinutes,
    d3_timeHours = d3.timeHours,
    d3_timeDays = d3.timeDays,
    d3_timeWeeks = d3.timeWeeks,
    d3_timeMonths = d3.timeMonths,
    d3_timeYears = d3.timeYears,
    // formatter
    d3_format = d3.format,
    d3_timeFormat = d3.timeFormat,
    d3_timeFormatUTC = d3.utcFormat,
    d3_timeParse = d3.timeParse,
    d3_formatLocale = d3.formatLocale,
    d3_timeFormatLocale = d3.timeFormatLocale,
    // d3_ajax
    d3_json = d3.json,
    // timer
    d3_timer = d3.timer,
    // brush
    d3_brush = d3.brush,
    d3_brushX = d3.brushX,
    d3_brushY = d3.brushY,
    d3_brushSelection = d3.brushSelection,
    // geo

    topojson = d3.topojson,
    versor = d3.versor,
    d3_geoPath = d3.geoPath,
    d3_geoGraticule = d3.geoGraticule,
    d3_geoCentroid = d3.geoCentroid,
    d3_polygonContains = d3.polygonContains,
    d3_geoDistance = d3.geoDistance,
    d3_geoContains = d3.geoContains,
    d3_quadtree = d3.quadtree,
    //Namespace
    d3_namespace = d3.namespace,
    d3_namespaces = d3.namespaces,
    //transition
    d3_transition = d3.transition,
    d3_customEvent = d3.customEvent,
    //events
    d3_event = function (e) {
        return arguments.length ? (d3_zc.event = e, e) : d3_zc.event;
    },
    // sankey variables
    d3_linkHorizontal = d3_zc.linkHorizontal,
    d3_polygonCentroid = d3_zc.polygonCentroid,
    d3_path = d3_zc.path;
