// @zohocharts/zc/src/core/data-parser.js
//$Id$
/*
 * this file contains methods to extract the required forms of data from the user input
 */
import { Registry } from './zc-registry';
import { CUSTOM_VALUE_PATTERN, charttype, stack } from './main-zc-init';
import { defined, isUndefined, decodeHTML, encodeHTML, isArray, getPropVal, simpleClone, pick, arrayIncludes, splat, getLevel, getDepth, isNumber } from '../utils/zcutils';
import { ORDINAL_DATATYPE, NUMERIC_DATATYPE, TIME_DATATYPE } from './datatype-constants';
import { DATALABELS_LABELTYPE, INNERLABEL_LABELTYPE, PARENTLABELS_LABELTYPE, STACKLABELS_LABELTYPE } from '../label/labeltype-constants';
import { PER_SHOWAS, MAXPER_SHOWAS, SPLITPER_SHOWAS } from '../label/showas-constants';
import { d3_merge } from './d3-constants';
import { mathAbs } from './zc-constants';
import { colorscale_helpers } from '../axis/colorscale-helpers';
import { isMultiPieChartType } from '../events/events-utils';
import { getFormattedValue, getFormattedValueByPercent } from '../formatter/formatter';
import { bubblepieUtils } from '../bubblepie/bubblepie-utils';
import { DataProcessor } from './data-processor';
// import { getStackedTotal } from './data-processor/stackedTotal';
import { isFacet } from '../axis/hierarchy-axis-utils';
import { SANKEY_CHARTNAME } from './chartname-constants';
import { getCurrentRendererIndex } from './data-processor/pivot-layout-helpers';
import { getHTML4SVGSymbol, getSVGSymbolString } from '../utils/svg-symbols';

export var dataparser = function () { };
Registry.setComponent('defaultDataParser', dataparser);

dataparser.get = function (data, chartObj) {
    var dataSplicer = chartObj.dataSplicer,
        metadata = data.metadata,
        columns = dataSplicer.getColumns(),
        axes = dataSplicer.getEncode();
    var seriesdata = data.seriesdata.chartdata;

    // data processor helpers
    var dataProcessorHelpers = DataProcessor.helpers,
        getParsedYAxisOrder = dataProcessorHelpers.getParsedYAxisOrder;

    axes = sortAxesObject();
    var dataObject,
        columnInfo_clr = columns[axes.clr] || {};
    var colTypeOrder = columns.map(function (d, i) {
        return getColTypeOrder(i);
    });

    var dataset = {
        xidx: columns[axes.x] && columns[axes.x].dataindex,
        yidx: function (order) {
            var columnidx = order instanceof Array ? axes.y[order[0]][order[1] || 0] : axes.y[order][0];
            return columns[columnidx].dataindex;
        },
        /**
         * 
         * @param {*} order - level order
         * @returns dataindex of the respective level
         */
        levelidx: function (order) {
            var columnidx = dataset.getColumnIdx('level', order);
            return columns[columnidx].dataindex;
        },
        zidx: (columns[axes.z] || {}).dataindex,
        shapeidx: (columns[axes.shape] || {}).dataindex,
        tooltipidx: function (index) {
            return axes.tooltip != null ? columns[index].dataindex : null;
        },
        clridx: columnInfo_clr.levelindex ? null : columnInfo_clr.dataindex
    };
    dataset.columnMappedByName = {}

    columns.map((d,i) => {
        dataset.columnMappedByName[d.columnname] = d;
    })
    dataset.setDataObj = function (obj) {
        dataObject = obj;
    };
    (dataset.getX = function (d) {
        if (
            defined(dataObject) &&
            ((dataObject.categoryHierarchy &&
                dataObject.hierarchical.pivotLayout.columns.count &&
                !dataObject.hierarchical.pivotLayout.columns.splitColumnLeaves) ||
                dataObject.isWaterfallEnabled)
        ) {
            //TODO waterfall
            return d.index;
        } else {
            return this.getActualX(d);
        }
    }),
        (dataset.getY = function (d, yaxiscolumnorder, range) {
            /*
             * Meta data for y should be in array
             * Axis order may be array of objects or an object
             */
            try {
                var curYidx = this.yidx(yaxiscolumnorder);
                var yVal = d[curYidx] === undefined ? null : range == null ? d[curYidx] : d[curYidx][range];

                return yVal;
                // eslint-disable-next-line no-empty
            } catch (err) { }
        }),
        (dataset.getZ = function (d) {
            return d[this.zidx];
        }),
        (dataset.getClr = function (d) {
            return d[this.clridx];
        }),
        (dataset.getGroupByValues = function (d, indices) {
            if (indices == null || indices.length == 0) {
                return null;
            }
            return indices.reduce(function (prev, cur, i) {
                return defined(d[cur]) ? prev + d[cur] : prev;
            }, '');
        });

    dataset.getActualX = function (d) {
        return d[this.xidx];
    };

    dataset.getDataType = function (type, index, seriestypes) {
        var dataSplicer = chartObj.dataSplicer,
            metadata_axes = dataSplicer.getEncode();

        seriestypes =
            seriestypes ||
            dataset.seriesTypes.values().map(function (d) {
                return +d;
            });
        let hierarchical = dataObject && dataObject.hierarchical,
            isBranchType;
        if (hierarchical) {
            if (hierarchical.pivot) {
                let tag = type == 'x' ? 'column' : 'row',
                    nodeList = dataObject.hierarchical.pivotLayout.getRootNodeList(tag),
                    leavesCount = nodeList.getLeavesCount();
                isBranchType = index > leavesCount - 1;
            } else {
                isBranchType = type === 'x' && index > 0;
            }
        }
        if (isBranchType) {
            return 'ordinalNonLinear';
        }

        var datatype = null;
        if (metadata_axes[type] != null) {
            var columnidx = this.getColumnIdx(type, index);
            var columnInfo = columns[columnidx];
            datatype = columnInfo.datatype;
            //ordinal cases:
            var ordinalType =
                (type == 'x' &&
                    ((dataObject && dataObject.isPolarAxisCategory) ||
                        seriestypes.indexOf(charttype.bar) != -1 ||
                        seriestypes.indexOf(charttype.stkdbar) != -1 ||
                        seriestypes.indexOf(charttype.bullet) != -1 ||
                        seriestypes.indexOf(charttype.gantt) != -1 ||
                        seriestypes.indexOf(charttype.boxplot) != -1 ||
                        seriestypes.indexOf(charttype.barrange) != -1 ||
                        seriestypes.indexOf(charttype.web) != -1)) ||
                (type === 'clr' && data.legend._dummy); //ZC2635

            var isHeatmap = seriestypes.indexOf(charttype.heatmap) != -1 && (type === 'x' || type === 'y');
            datatype = ordinalType || columnInfo.useDiscrete || isHeatmap ? ORDINAL_DATATYPE : datatype;
        }
        return datatype;
    };

    dataset.getColumnIdx = function (type, order) {
        var columnidx;
        var dataSplicer = chartObj.dataSplicer,
            metadata_axes = dataSplicer.getEncode(),
            metadata_columns = dataSplicer.getColumns();
        var axes = metadata_axes[type];
        if (type == 'y') {
            if (dataObject && dataObject.hierarchical && dataObject.hierarchical.pivotLayout.rows.count) {
                let rowList = dataObject.hierarchical.pivotLayout.getRootNodeList('row');
                if (order > rowList.getLeavesCount() - 1) {
                    metadata_columns.filter(function (d, i) {
                        if (d.levelindex == chartObj.axes.y.count - 1 - order) {
                            columnidx = i;
                        }
                    });
                } else {
                    columnidx = axes[0][0];
                }
            } else {
                columnidx = order instanceof Array && order.length == 2 ? axes[order[0]][order[1]] : axes[order][0];
            }
        } else if (type == 'tooltip') {
            columnidx = axes[order];
        } else if (type == 'x' && order > 0) {
            if (dataObject && dataObject.hierarchical && dataObject.hierarchical.pivot) {
                let columnList = dataObject.hierarchical.pivotLayout.getRootNodeList('column'),
                    rowList = dataObject.hierarchical.pivotLayout.getRootNodeList('row'),
                    rowCount = rowList.getBranchesCount() || 0;
                if (order > columnList.getLeavesCount() - 1) {
                    metadata_columns.filter(function (d, i) {
                        if (d.levelindex == rowCount + chartObj.axes.x.count - 1 - order) {
                            columnidx = i;
                        }
                    });
                } else {
                    columnidx = axes[0];
                }
            } else {
                metadata_columns.filter(function (d, i) {
                    if (d.levelindex == chartObj.axes.x.count - order) {
                        columnidx = i;
                    }
                });
            }

            columnidx = pick(columnidx, axes[0]);
        } else if (type == 'level') {
            let order_0 = order && order.length > 1 ? order[0] : 0,
                order_1 = !isNumber(order) ? order.length > 1 ? order[1] : order[0] : order;

            order_1 = order_1 < 0 ? axes.length - 1 : order_1;
            columnidx = isArray(axes[order_0]) ? axes[order_0][order_1] : axes[order_1];
        } else {
            columnidx = axes == null ? null : axes[0];
        }
        return columnidx;
    };

    dataset.getColumn = function (type, yaxiscolumnorder, columnName) {
        if(defined(columnName)){
            return this.columnMappedByName[columnName];
        }
        var columnidx = this.getColumnIdx(type, yaxiscolumnorder);
        return columns[columnidx];
    };
    dataset.getScaleType = function (type, yaxiscolumnorder, columnName) {
        const column = dataset.getColumn(type, yaxiscolumnorder, columnName) || {};
        return (column.scaleType || column[column.datatype]?.scaleType || '').toLowerCase();
    }

    dataset.getColumnName = function (colidx, toDecode) {
        return (toDecode != false ? decodeHTML : encodeHTML)(columns[colidx].columnname);
    };

    /**
     * It is used to find the levelindex of the respective series based on first occurence of column index
     * @param {number} colidx - column index
     * @param {*} order - level order
     * @returns index of the first occurence of column index in metadata.level[order] array
     */
    dataset.getLevelIndex = function (colidx, order = 0) {
        let levelAxes = chartObj.dataSplicer.getEncode().level;
        levelAxes = isArray(levelAxes, 2) ? levelAxes[order] : levelAxes || [];
        return levelAxes.findIndex(_ => _ === colidx);
    }

    dataset.getCustomizedVal = function (
        chartObj,
        colidx,
        seriespos,
        subseriespos,
        itempos,
        d,
        type,
        color,
        toDecode,
        subseriesLength,
        yrangePos
    ) {
        if (isUndefined(toDecode)) {
            toDecode = !/(tooltip|legend)/.test(type);
        }

        let isSankey = chartObj.chartTypes[0].name === SANKEY_CHARTNAME;

        return colidx.toString().replace(CUSTOM_VALUE_PATTERN, function () {
            var matchedString = arguments[0],
                valIndex = arguments[1], //INDEX value of {{val(INDEX)}}
                colIndex = arguments[2], //INDEX value of {{col(INDEX)}}
                perIndex = arguments[3], //INDEX value of {{per(INDEX)}}
                splitPerIndex = arguments[4], //INDEX value of {{splitper(INDEX)}}
                maxPerIndex = arguments[5],
                valIndex_1 = arguments[7], //Value at array position
                dataindex = arguments[8]; //yrangepos
            const symbolConfigName = arguments[9], // symbol configuration name of {{symbol(symbolConfig)}}
                symbolColor = arguments[10]; // color of the symbol
            var idx = valIndex || colIndex || perIndex || splitPerIndex || maxPerIndex || dataindex;

            if (matchedString.indexOf('{{symbol(') !== -1) {
                // returns an image URL based on configuration.
                const symbolConfig = (symbolConfigName !== "null") ? Object.assign({}, $ZC.utils.getPropVal(window, symbolConfigName)) : {};
                if (symbolConfig === undefined) {
                    throw 'Symbol configuration not defined on window object';
                }
                if (!defined(symbolConfig.fillColor)) {
                    symbolConfig.fillColor = !defined(symbolColor) ? "black"
                        : (symbolColor != "{{clr}}") ? symbolColor
                            : (typeof color === "string") ? color
                                : color[colidx];
                }
                // Shape based on legend.shapes
                const shapes = chartObj.systemConf.legend.shapes;
                if (typeof shapes === 'string') {
                    symbolConfig.shape = symbolConfig.shape || shapes;
                } else {
                    const idx = d ? (!defined(d.seriesIndex) ? (!defined(d._seriesIndex) ? (!defined(d.seriesPos) ? itempos : d.seriesPos) : d._seriesIndex) : d.seriesIndex) : seriespos;
                    symbolConfig.shape = symbolConfig.shape || ((shapes && shapes.length > 0) ? shapes[idx % shapes.length] : 'circle');
                }

                symbolConfig.size = symbolConfig.size || 15;
                const width = symbolConfig.width || symbolConfig.size,
                    height = symbolConfig.height || symbolConfig.size,
                    pathString = getHTML4SVGSymbol(symbolConfig),
                    image = getSVGSymbolString(pathString, width, height, true);
                return image;
            }

            if (matchedString.indexOf('{{clr}}') != -1) {
                // theme color based on series color
                return color;
            }
            if (columns[idx] != null) {
                var dataPoint =
                    type === STACKLABELS_LABELTYPE || isSankey
                        ? d
                        : dataset.getDataPoint(chartObj, idx, seriespos, subseriespos, itempos, d);
                if (dataPoint != null) {
                    var operation = matchedString.match(/[a-z]+/gi)[0];
                    switch (operation) {
                        case 'col':
                            return dataset.getColumnName(colIndex, toDecode); //return columnname
                        case 'valatpos':
                        case 'val':
                            var isRangeDefined = defined(dataindex),
                                yArrayIndex = yrangePos;
                            if (isRangeDefined) {
                                (valIndex = valIndex_1), (yArrayIndex = dataindex);
                            }
                            var val = dataset.getDataPointValue(
                                chartObj,
                                dataPoint,
                                valIndex,
                                seriespos,
                                false,
                                type,
                                subseriespos,
                                toDecode,
                                subseriesLength,
                                yArrayIndex
                            );

                            if (
                                [DATALABELS_LABELTYPE, STACKLABELS_LABELTYPE, INNERLABEL_LABELTYPE, PARENTLABELS_LABELTYPE].indexOf(type) >
                                -1
                            ) {
                                val = val || ''; //Test_437.json
                            }
                            return val; //return value
                        case PER_SHOWAS:
                        case SPLITPER_SHOWAS:
                        case MAXPER_SHOWAS:
                            return dataset.getDataPointValue(
                                chartObj,
                                dataPoint,
                                perIndex || splitPerIndex || maxPerIndex,
                                seriespos,
                                operation,
                                type,
                                subseriespos,
                                toDecode,
                                subseriesLength
                            ); //return the percentage of value (percentage of Y value is only supported now)
                    }
                } else {
                    return '';
                }
            } else {
                return matchedString;
            }
        });
    };

    dataset.getYaxisColumnIndexBySeriesIndex = function (chartObj, seriesIndex, d) {
        var hierarchical = chartObj.dataObject.hierarchical,
            facetRendering = hierarchical && hierarchical.facetRendering,
            _seriesdata = (seriesdata[seriesIndex] || seriesdata[0]);

        if(facetRendering && d && d.paneid){
            var pivotLayout = hierarchical.pivotLayout,
                layerGrp = pivotLayout.layers.groups,
                layerPane = pivotLayout.layers.panes,
                layers = layerGrp[layerPane[d.paneid].data[seriesIndex].layer],
                legendIdx = layers.indexOf(layerPane[d.paneid].data[seriesIndex].value),
                hasMultiCluster = facetRendering && pivotLayout.rows && pivotLayout.rows.hasMultiCluster(),
                hasMultiLayers = hasMultiCluster ? pivotLayout.layers.isLayerNode : false;
            _seriesdata = hasMultiLayers ? layerPane[d.paneid].data[legendIdx].parent.node : layerPane[d.paneid].data[seriesIndex].node;
        }

        var yaxiscolumnorder = _seriesdata.yaxiscolumnorder || [0, 0];
        return isArray(yaxiscolumnorder) ? axes.y[yaxiscolumnorder[0]][yaxiscolumnorder[1] || 0] : axes.y[yaxiscolumnorder][0];
    };

    dataset.getDataPoint = function (chartObj, colidx, seriespos, subseriespos, itempos, d) {
        //seriesdataAfterUpdate refers seriesdata without null data,till now handled only in bar chart
        colidx = +colidx;
        var dataPoint = null,
            colInfo = columns[colidx] || {},
            type = colTypeOrder[colidx] && colTypeOrder[colidx].type,
            levelindex = colInfo.levelindex,
            curSeriesPos = seriespos,
            hierarchyCondition = true;
            

        //TODO: Work on sankey hierarchy data model tooltip like sunburst tooltip
        if (chartObj.isLevelAxisCategory && this.axesDefined(colidx, ['x', 'y', 'level'])) {
            levelindex = dataset.getLevelIndex(colidx);
            if (d._type === 'link') {
                dataPoint =  levelindex < getDepth(d.level) ? d : null;
            } else {
                dataPoint = levelindex == getDepth(d.level) - 1 ? d : null;
            }
            return dataPoint;
        }

        var yaxisColumnIndex = chartObj.dataObject.isLevelAxisCategory ? this.levelidx(0) :  this.getYaxisColumnIndexBySeriesIndex(chartObj, seriespos, d);
        if (type === 'y' && yaxisColumnIndex !== colidx) {
            //For Multi y charts, show the respective series only
            return dataPoint; // why we are returning null data point for multi y charts
        }
        if (chartObj.dataObject.isHierarchy) {
            let rendererIndex = getRendererIndex(d, chartObj);
            var getProcessedNodeFunc = chartObj.renderer[rendererIndex].getProcessedNode;
            var node = getProcessedNodeFunc ? getProcessedNodeFunc(d) : d,
                yaxiscolOrder = d.yaxiscolumnorder,
                yDimension = chartObj.systemConf.metadata.axes.y.length,
                yaxisOrderValue = d.yaxisOrder,
                yValue = chartObj.systemConf.metadata.axes.y;
            let facets = chartObj.userdata.facets;
            hierarchyCondition =
                (axes.clr && axes.clr[0] === colidx && node.depth === 1) ||
                (defined(yaxiscolOrder) ? (yValue[yaxiscolOrder[0]]) : ((yDimension > 1 && defined(yaxisOrderValue)) ? yValue[yaxisOrderValue] : yValue[0])).indexOf(colidx) !== -1 ||
                (node.children &&
                    (colInfo.levelindex === node.depth - 1 ||
                        (isArray(colInfo.levelindex) && colInfo.levelindex.indexOf(node.depth - 1) !== -1))) ||
                (!node.children && (colInfo.levelindex == null || colInfo.dataindex != null)) ||
                arrayIncludes((facets && facets.rows) || [], colidx) ||
                arrayIncludes((facets && facets.columns) || [], colidx);
        } else if (chartObj.dataObject.isFlowChartCategory) {
            dataPoint = d.point;
        }

        if ((curSeriesPos != null && hierarchyCondition)) {
            let rendererIndex = getRendererIndex(
                d,
                chartObj,
                chartObj.isAxisCategory || chartObj.dataObject.isPolarAxisCategory ? curSeriesPos : 0
            );
            var renderer = chartObj.renderer[rendererIndex];
            var seriesdataAfterUpdate = d
                ? null
                : renderer != null && renderer.seriesdataAfterUpdate != null
                    ? renderer.seriesdataAfterUpdate[subseriespos]
                    : seriesdata[curSeriesPos].data[subseriespos];
            dataPoint = d || seriesdataAfterUpdate[itempos];
        }
        if(chartObj.customVisual && d){
            return d;
        }
        return dataPoint;
    };

    dataset.getDataPointValue = function (
        chartObj,
        dataPoint,
        colidx,
        seriespos,
        percent,
        type,
        subseriespos,
        toDecode,
        subseriesLength,
        yrangePos
    ) {
        if (colidx != null) {
            colidx = +colidx;
        }

        if (isUndefined(toDecode)) {
            toDecode = !/(tooltip|legend)/.test(type);
        }

        var colInfo = columns[colidx] || {},
            datatype = colInfo.datatype,
            levelindex = colInfo.levelindex,
            dataindex = colInfo.dataindex,
            node;
        var format = colInfo[datatype];
        if (dataPoint != null) {
            if (chartObj.dataObject.isFlowChartCategory) {
                if (chartObj.isLevelAxisCategory && this.axesDefined(colidx, ['x', 'y', 'level'])) {
                    const depth = getDepth(dataPoint.level) - 1;
                    levelindex = dataset.getLevelIndex(colidx);
                    dataindex = dataPoint._type == 'node' || depth != levelindex ? 0 : 1;
                    if (type === 'tooltip') {
                        dataPoint = depth > levelindex ? getParent(dataPoint, levelindex) : dataPoint;
                    }
                }
            }
            if (chartObj.dataObject.isHierarchy) {
                let rendererIndex = getRendererIndex(dataPoint, chartObj);
                var renderer = chartObj.renderer[rendererIndex];
                var getProcessedNodeFunc = renderer && renderer.getProcessedNode;
                node = getProcessedNodeFunc ? getProcessedNodeFunc(dataPoint) : dataPoint;
            }

            if (
                (axes.clr != null || type === 'legend') &&
                colidx == axes.clr &&
                !(chartObj.dataObject.isHierarchy && type !== 'legend' && colInfo.dataindex != null)
            ) {
                //Eventhough clrIdx is not specified in metadata, the seriesname will be returned correctly
                if (defined(this.clridx) && colorscale_helpers.isQuantileColorScale(chartObj)) {
                    if (type === 'legend') {
                        var arraySeperator = '-';
                        return getFormattedValue(format, dataPoint, datatype, chartObj, type, toDecode, arraySeperator);
                    }
                    return getFormattedValue(format, this.getClr(dataPoint), datatype, chartObj, type, toDecode);
                } else {
                    var actualData = dataPoint.paneid
                        ? DataProcessor.helpers.getSeriesdata(chartObj, dataPoint.paneid, seriespos)
                        : seriesdata[seriespos];
                    return getFormattedValue(
                        format,
                        dataPoint.seriesname || (actualData && actualData.seriesname),
                        datatype,
                        chartObj,
                        type,
                        toDecode
                    );
                }
            } else if (defined(levelindex) && chartObj.dataObject.hierarchical && datatype == 'ordinal') {
                var nodeValue = chartObj.dataObject.hierarchical.getNodeValueByColumnIdx(dataPoint, colidx, levelindex);

                if (nodeValue) {
                    return getFormattedValue(format, nodeValue, datatype, chartObj, type, toDecode);
                } else {
                    return '';
                }
            } else if (
                chartObj.dataObject.isHierarchy &&
                (levelindex === node.depth - 1 || (isArray(levelindex) && levelindex.indexOf(node.depth - 1) !== -1))
            ) {
                var val = !node.children && colInfo.dataindex != null ? dataPoint[dataindex] : dataPoint.seriesname;
                return getFormattedValue(format, val, datatype, chartObj, type, toDecode);
            } else {
                var value;
                var yidx = chartObj.isLevelAxisCategory ? this.getColumnIdx('level', 0) :
                 this.getColumnIdx( 'y', getParsedYAxisOrder((chartObj.seriesdata[seriespos] || chartObj.seriesdata[0]).yaxiscolumnorder)
                );
                var arraySeperator = colidx == yidx || chartObj.dataObject.hasWhiskers ? ' - ' : undefined;
                var formatFn = percent ? getFormattedValueByPercent : getFormattedValue;
                if (subseriesLength) {
                    //subserieslength - to iterate all subseries y value to show for tooltip shared series info
                    var cumval = 0;
                    for (var j = 0; j < subseriesLength; j++) {
                        // var dp = dataparser.getDataPoint(chartObj, yidx, seriespos, j, itempos);
                        cumval += dataset.getDataPointUnformattedValue(
                            datatype,
                            seriespos,
                            j,
                            colidx,
                            dataPoint,
                            type,
                            percent,
                            dataindex,
                            yrangePos
                        );
                    }
                    value = cumval;
                } else {
                    value = dataset.getDataPointUnformattedValue(
                        datatype,
                        seriespos,
                        subseriespos,
                        colidx,
                        dataPoint,
                        type,
                        percent,
                        dataindex,
                        yrangePos
                    );
                }

                if (value === '' && percent) {
                    return value;
                }
                return formatFn(format, value, datatype, chartObj, type, toDecode, arraySeperator);
            }
        }
        return null;
    };
    dataset.getDataPointUnformattedValue = function (
        datatype,
        seriespos,
        subseriespos,
        colidx,
        dataPoint,
        type,
        percent,
        dataindex,
        yrangePos
    ) {
        var value;
        if (chartObj.dataObject.isHierarchy) {
            let rendererIndex = getRendererIndex(dataPoint, chartObj);
            var renderer = chartObj.renderer[rendererIndex];
            var getProcessedNodeFunc = renderer && renderer.getProcessedNode,
                node = getProcessedNodeFunc ? getProcessedNodeFunc(dataPoint) : dataPoint;
        }
        if (percent) {
            var total = datatype == NUMERIC_DATATYPE ? getTotal(chartObj, seriespos, subseriespos, colidx, dataPoint, percent, type) : null;
            if (total != null) {
                value =
                    ((type == STACKLABELS_LABELTYPE
                        ? dataPoint
                        : chartObj.dataObject.isHierarchy && node.children && defined(node.value)
                            ? node.value
                            : chartObj.dataObject.isPercentChart
                                ? dataPoint.y
                                : dataPoint[dataindex]) /
                        total) *
                    100;
                value = value || 0; //if total=0 and y=0, the 'val' holds NaN
                return value;
            } else {
                return '';
            }
        }
        
        var yidx = chartObj.isLevelAxisCategory ? this.getColumnIdx('level', 0) :
            this.getColumnIdx('y', (chartObj.seriesdata[seriespos] || chartObj.seriesdata[0]).yaxiscolumnorder || 0);
        if (chartObj.isLevelAxisCategory) {
            const levelindex = dataset.getLevelIndex(colidx);
            const xidx = dataset.levelidx(-1);
            value = levelindex >= 0 && isArray(dataPoint[xidx]) ? dataPoint[xidx][dataindex] : dataPoint[dataindex];
        } else {
            value =
                chartObj.dataObject.isHierarchy && node.children && colidx == yidx && defined(node.value)
                    ? node.value
                    : type == STACKLABELS_LABELTYPE
                        ? dataPoint
                        : defined(yrangePos)
                            ? dataPoint[dataindex][yrangePos]
                            : dataPoint[dataindex];
        }
        return value;
    };

    dataset.getMetaInfoObj = function (chartObj, data, seriespos) {
        var axes = chartObj.systemConf.metadata.axes,
            columns = chartObj.systemConf.metadata.columns,
            seriesdata = chartObj.systemConf.seriesdata.chartdata,
            seriespos = chartObj.isAxisCategory || chartObj.isNonAxisMultiSeriesCategory ? seriespos : 0,
            dataWithMetadata = [];
        
        const isLevelAxisCategory = chartObj.dataObject.isLevelAxisCategory;
        var yindex = isLevelAxisCategory ? [0, 0] : (seriesdata[seriespos] || seriesdata[0]).yaxiscolumnorder || [0, 0]; //yaxiscolumnorder is not mandatory
        const yaxes = isLevelAxisCategory ? isArray(axes.level, 2) ? axes.level: [axes.level] : axes.y;
        var yColIdx = yindex instanceof Array ? yaxes[yindex[0]][yindex[1] || 0] : yaxes[yindex][0];
        if (data instanceof Array) {
            data = isLevelAxisCategory ? (data.original || data) : data;
            data.forEach(function (d, i) {
                let colidx;
                if (chartObj.dataObject.isLevelAxisCategory && isArray(d)) {
                    let depth = getDepth(data.level) - 2;
                    colidx = d.map(_ => {
                        return dataset.getColumnIdx('level', depth++);
                    })
                } else {
                    colidx = columns.indexOf(
                        columns.filter(function (c, j) {
                            if (isLevelAxisCategory && dataset.levelidx(-1) === i && dataset.axesDefined(j, 'level')) {
                                const depth =  getDepth(data.level) - 1;
                                return c.dataindex === dataset.levelidx(depth) && dataset.getColumnIdx('level', depth) === j;
                            } else {
                                return isLevelAxisCategory ? (c.dataindex === i)
                                    : c.dataindex == i && (dataset.yidx(yindex) != i || (dataset.yidx(yindex) == i && yColIdx == j));
                            }
                        })[0]
                    );
                }
                dataWithMetadata.push(getColObj(colidx, d));
            });
        }

        if (axes.clr != null) {
            var clrColIdx = axes.clr instanceof Array ? axes.clr[0] : axes.clr;
            dataWithMetadata.push(getColObj(clrColIdx, seriesdata[seriespos].seriesname));
        }
        return dataWithMetadata;

        function getColObj(colidx, value) {
            if (isArray(colidx)) {
                return colidx.map((_, i) => getColObj(_, value[i]));
            }
            var obj = simpleClone(columns[colidx] || {});
            obj.value = value;
            obj.columnindex = colidx;
            return obj;
        }
    };

    dataset.processMetaData = function () {
        var legendIndex = getPropVal(metadata, 'axes.clr');

        if (defined(legendIndex)) {
            legendIndex = legendIndex[0];
        }

        function process(metadata, i) {
            var type;

            if (i === legendIndex) {
                return;
            }

            if (!defined(metadata.datatype)) {
                var data = dataset.getDataByColumnIndex(metadata.dataindex);
                type = dataset.findDataType(data);
                metadata.datatype = type;
            }
            if (defined(type) && !defined(metadata[type])) {
                metadata[type] = {};
            }
        }
        columns.forEach(process);
    };

    dataset.getDataByColumnIndex = function (index) {
        var data = null,
            found = false;
        chartObj.seriesdata.some(function (series) {
            return series.data.some(function (subSeries) {
                return subSeries.some(function (point) {
                    if (defined(point)) {
                        found = defined(point[index]);
                        if (found) {
                            data = point[index];
                        }
                        return found;
                    }
                });
            });
        });
        return data;
    };

    dataset.findDataType = function (data) {
        if (typeof data === 'number') {
            if (data > 365 * 24 * 3600 * 1000) {
                return TIME_DATATYPE;
            }
            return NUMERIC_DATATYPE;
        }
        return ORDINAL_DATATYPE;
    };

    // init
    dataset.processMetaData();
    // handle meta data.

    function getColTypeOrder(colidx) {
        var axesType,
            order = [];
        for (var type in axes) {
            if (axes.hasOwnProperty(type)) {
                var val = type == 'y' ? d3_merge(axes[type]) : axes[type];
                if (val.indexOf(colidx) != -1) {
                    if (type == 'y') {
                        axes[type].map(function (d, i) {
                            var index = d.indexOf(colidx);
                            if (index != -1) {
                                order.push(i);
                                order.push(index);
                            }
                        });
                    }
                    axesType = type;
                    return {
                        type: axesType,
                        order: order
                    };
                }
            }
        }
        return {};
    }

    /**
     * Checks whether the provided axesType is defined in columns
     * @param {number} colidx - column index in metadata.columns
     * @param {Array[String]} axesType - axes type in metadata.axes
     * @returns {Boolean} whether the column is defined or not in given axes type
     */
    dataset.axesDefined = function (colidx, axesType) {
        return splat(axesType).includes(colTypeOrder[+colidx] && colTypeOrder[+colidx].type);
    }

    function sortAxesObject() {
        var newaxes = {},
            keysSorted = Object.keys(axes).sort(function (a, b) {
                return b.length > a.length ? -1 : b.length < a.length ? 1 : 0; // (a.length > b.length) is incorrect in SAFARI
            });
        keysSorted.forEach(function (d) {
            newaxes[d] = axes[d];
        });
        return newaxes;
    }

    return dataset;
};

export function getTotal(chartObj, seriespos, subseriespos, colidx, dataPoint, showAs, type) {
    colidx = +colidx;
    var axes = chartObj.systemConf.metadata.axes,
        yaxes = axes.y,
        index = -1,
        isBubblePie = isMultiPieChartType(chartObj.chartTypes[seriespos].value),
        stacked = chartObj.dataObject.isstacked(chartObj.chartTypes[seriespos].value);
    if (stacked && !chartObj.dataObject.isPercentChart && /(datalabels|tooltip)/.test(type)) {
        return stack.getStackedTotal(chartObj, seriespos, dataPoint, showAs);
    }

    if (chartObj.dataObject.isFlowChartCategory && chartObj.dataObject.hasHierarchyData) {
        const { getTopLevel, getTotalValue } = chartObj.dataObject.sankeyHierarchyLayout.utils;  // percentage value support in hierarchical sankey chart
        return (showAs == SPLITPER_SHOWAS && dataPoint._type == 'link') ? getTotalValue(getTopLevel(dataPoint.level)) : getTotalValue();
    }

    if (isArray(yaxes) || chartObj.isGeoMap) {
        if (isArray(yaxes)) {
            if (isArray(yaxes[0])) {
                yaxes.forEach(function (d, i) {
                    index = index == -1 && d.indexOf(colidx) > -1 ? i : index;
                });
            } else {
                index = yaxes.indexOf(colidx) > -1 ? 0 : -1;
            }
        }

        if (index != -1 || isBubblePie) {
            var total;
            if (isBubblePie) {
                var key = bubblepieUtils.getKey(dataPoint, seriespos, chartObj);
                var pieRenderer = chartObj.subrenderer[subseriespos || 0][key];
                total = pieRenderer.total;
            } else if (chartObj.isAxisCategory || chartObj.dataObject.isPolarAxisCategory) {
                var yscale = chartObj.axes.y.groups[index].scale;
                if (showAs === MAXPER_SHOWAS || !yscale) {
                    var domain = chartObj.dataObject.ZC_BAR_getRangeY(index);
                    total = chartObj.dataset.getY(dataPoint, index) < 0 ? -domain[0] : domain[1];
                } else {
                    var domain = yscale.domain();
                    total = domain[1];
                }
            } else {
                if (chartObj.isNonAxisMultiSeriesCategory) {
                    let rendererIndex = getRendererIndex(dataPoint, chartObj);
                    var renderer = chartObj.renderer[rendererIndex];
                    if (showAs === PER_SHOWAS) {
                        return renderer.total;
                    } else if (renderer.extentY) {
                        return renderer.extentY[1];
                    } else {
                        var getProcessedNodeFunc = renderer.getProcessedNode;
                        var node = getProcessedNodeFunc ? getProcessedNodeFunc(dataPoint) : dataPoint;
                        total = node._parent ? node._parent.value : node.parent.value;
                    }
                } else if (type === 'legend') {
                    var curSeriesData = chartObj.seriesdata[0],
                        yDataIndex = chartObj.dataset.yidx(DataProcessor.helpers.getParsedYAxisOrder(curSeriesData.yaxiscolumnorder));
                    total = curSeriesData.data[0].reduce(function (prev, cur) {
                        var yVal = cur.disabled ? 0 : mathAbs(cur[yDataIndex] || 0);
                        return prev + yVal;
                    }, 0);
                } else {
                    let rendererIndex = getRendererIndex(dataPoint, chartObj);
                    total = chartObj.renderer[rendererIndex].total;
                }
            }
            return !total && total !== 0 ? null : total;
        }
    }
    return null;
}

export function getRendererIndex(d, chartObj, seriesIndex = 0) {
    if (isFacet(chartObj) && ((d && d.data && d.data.paneid) || (d && d.paneid))) {
        let panes = chartObj.dataObject.hierarchical.pivotLayout.getRootNodeList('layer').panes,
            paneIndex = panes[d.paneid].index;

        return getCurrentRendererIndex(panes, paneIndex);
    }

    return seriesIndex;
}

function getParent(data, levelindex) {
    let depth = getDepth(data.level) - levelindex - 1;
    while (--depth > 0) {
        data = data._root.parent.point;
    }
    return data;
}
