// @zohocharts/zc/src/core/data-processor.js
//data-processor.js
/**
 *
 * @param data
 * @param dataSet
 * @param chartInstance
 * @constructor
 */

import { d3_set } from './d3-constants';
import { isArray, extend, arrayIncludes, isUndefined } from '../utils/zcutils';
import { NUMERIC_DATATYPE } from './datatype-constants';
import { charttype, chartcategory, charttypenames } from './main-zc-init';
import { Registry } from './zc-registry';
import { encodeMethods } from './data-processor/encode-helpers';
import { getGridList } from './data-processor/pivot-layout-helpers';
import { POLAR_SCALETYPE } from '../scale/scale-type-constants';

Registry.setComponent('defaultDataProcessor', DataProcessor);
export function DataProcessor(data, dataSet, chartInstance) {
    this.data = data;
    this.dataset = dataSet;
    this.chartObj = chartInstance;

    this.allChartTypes = null;
    this.globalChartType = null;
    this.processedseriesData = this.sankeyHierarchyLayout ? [{data: [this.sankeyHierarchyLayout.completeGraph.links]}] : chartInstance.seriesdata;

    this.initialize();
}

var DataProcessorProto = DataProcessor.prototype;

/**
 *
 * @param state
 */

DataProcessorProto.initialize = function(state) {
    var processor = this,
        data = processor.data,
        globalChartType = (processor.globalChartType = data.seriesdata && data.seriesdata.type),
        metadata = data.metadata,
        dataset = processor.dataset,
        processedseriesData = processor.processedseriesData,
        helpers = DataProcessor.helpers,
        chartObj = processor.chartObj;
    
    processor.globalChartType = processor.globalChartType || (chartObj.customVisual ? 'bar' : null);

    dataset.setDataObj(processor);
    dataset.seriesTypes = d3_set();
    processor.totalnoofdatas = 0;
    processor.encodes = {};
    processor.markerSize = {};
    processor.annotation = [];

    chartObj.cache.data = {};
    chartObj.cache.search = {};
    chartObj.cache.customPoints = {};
    chartObj.cache.correction = {};

    if (isArray(processedseriesData)) {
        var groupbyDataIndices = helpers.groupByDataIndices(metadata, chartObj);
        processor.allChartTypes = processor.allChartTypes || helpers.getAllChartTypes(data);

        let gridlist = processor.pivot ? getGridList(processor) : [{ data: processedseriesData }];
        gridlist.forEach(grid => {
            grid.data.forEach(d => {
                var chartType_num = helpers.getChartType((d.node && d.node.type) || d.type || globalChartType, NUMERIC_DATATYPE);
                const xScaleType = dataset.getColumn('x')?.scaleType;
                // isAxisCategory, isPolarAxisCategory, isNonAxisMultiSeriesCategory should be here for totalDataCount (DEMOEVENTS_BUBBLEPIE_4)
                
                processor.isAxisCategory = processor.isAxisCategory || (arrayIncludes(chartcategory.axis, chartType_num) && xScaleType != POLAR_SCALETYPE);

                processor.isPolarAxisCategory = processor.isPolarAxisCategory || arrayIncludes(chartcategory.polarAxis, chartType_num) || xScaleType == POLAR_SCALETYPE;
                processor.isNonAxisMultiSeriesCategory =
                    processor.isNonAxisMultiSeriesCategory || arrayIncludes(chartcategory.nonaxismultiseries, chartType_num);

                processor.singleSourceEvent = processor.singleSourceEvent || arrayIncludes(chartcategory.singleSourceEvent, chartType_num);

                if(chartObj.viewFlow && chartObj.viewFlow.spec.renderOldConfig == false){
                     processor.isAxisCategory = chartObj.viewFlow.spec.axes && chartObj.viewFlow.spec.axes.length;
                     processor.singleSourceEvent = false;
                }

                processor.hasLevelMarker = processor.hasLevelMarker || chartType_num == charttype.bullet || chartType_num == charttype.dial;

                // Initiate markerSize
                if (chartcategory.markerPlot.has(chartType_num)) {
                    processor.markerSize[charttypenames.get(chartType_num)] = d3_set();
                }
            });
        });

        processedseriesData.forEach(function(d, i) {
            var disabled = d.disabled,
                curdata = d.data,
                is3DArray = helpers.is3D(curdata),
                chartType_num = helpers.getChartType(d.type || globalChartType, NUMERIC_DATATYPE);

            dataset.seriesTypes.add(chartType_num);

            // convert data to 3d array even it is disabled;
            if (state !== 'addorremoveseries' && !is3DArray && isArray(curdata)) {
                curdata = processor.convertTo3DArray(d, groupbyDataIndices, chartType_num);
            }

            if (!disabled && isArray(curdata)) {
                //totalNo of datas
                processor.totalDataCount(curdata);
            }
            saveAnnotation(curdata, chartObj, processor.annotation, i);
        });

        // findSubFunction(processor, processedseriesData);
    }
};

function saveAnnotation(curdata, chart, arr, seriesIndex) {
    let config = chart.systemConf.notes || {},
        { dataIndex } = config;

    if (isUndefined(dataIndex)) {
        return;
    }

    curdata.forEach((subgroup, subSeriesIndex) => {
        if (isArray(subgroup)) {
            subgroup.forEach((d, itemIndex) => {
                let data = d.data || d;
                if (data[dataIndex]) {
                    arr.push({ data: d, subSeriesIndex, itemIndex, seriesIndex });
                }
            });
        }
    });
}

extend(DataProcessorProto, encodeMethods);
