// @zohocharts/zc/src/core/defaulttheme.js
//$Id$
//defaulttheme.js

import { mergeJSON } from '../utils/zcutils';
import { mathPI } from './zc-constants';
import { $ZCG } from './zcglobal';

export var defaultTheme = function(userProp) {
    //userProp for getting fontSize wrt chartObj not globally
    var defaulttheme = {
        canvas: {
            fontFamily: $ZCG.globalProp.fontL1Family, //Note: import font to your application if it is not supported by the browser.
            fontColor: $ZCG.globalProp.fontL1Color,
            fontSize: userProp.fontL1Size.l3,
            fontWeight: $ZCG.globalProp.fontL1Weight, //normal || bold || bolder || lighter || 100 || 200 || 300 || 400 || 500 || 600 || 700 || 800 || 900
            fontStyle: $ZCG.globalProp.fontL1Style, //normal || italic || oblique

            title: {
                show: true,
                text: 'Title comes here',
                hAlign: 'left',
                vAlign: 'top',
                marginTop: 10,
                marginRight: 0,
                marginBottom: 0,
                marginLeft: 10,
                target: 'newtab',
                fontSize: userProp.fontL1Size.l1
            },
            subtitle: {
                show: true,
                text: 'Description comes here',
                hAlign: 'left',
                vAlign: 'top',
                marginTop: 0,
                marginRight: 0,
                marginBottom: 0,
                marginLeft: 10,
                target: 'newtab',
                fontSize: userProp.fontL1Size.l2
            },
            background: {
                alpha: 1,
                color: '#FFFFFF'
            },

            border: {
                show: true,
                size: 1,
                style: 'solid',
                radius: 1,
                color: '#cccccc'
            },

            shadow: {
                show: true,
                x: 1,
                y: 1,
                color: 'rgba(0,0,0,0.1)',
                blur: 2,
                inset: false
            },
            intelligence: {
                dimension: {
                    title: 100, // 0 or null -> don't remove
                    subtitle: 120, // 0 or null -> don't remove
                    xaxis: 70, // 0 or null -> don't remove
                    xaxislabel: 80,
                    yaxis: 70, // 0 or null -> don't remove
                    yaxislabel: 80,
                    legend: 140, // 0 or null -> don't remove
                    credits: 150, // 0 or null -> don't remove
                    marker: 50
                },
                dataCount: {
                    marker: 300, //null -> don't remove
                    animation: 300, //null -> don't remove
                    datalabels: 300 //null -> don't remove
                }
            }
        },
        legend: {
            enabled: true,
            colorPallete: {
                type: 'multicolor', //muticolor || monochrome
                options: {
                    multicolor: 'countries',
                    monochrome: {
                        baseColor: 'red',
                        count: 5,
                        colorGamma: [0.3, 0.9]
                    }
                }
            },
            hAlign: 'right', //left, right, center
            vAlign: 'top', //top, center, bottom.
            fontSize: userProp.fontL1Size.l5,
            textPadding: 5,
            itemMargin: 3,
            marginTop: 10,
            marginRight: 10,
            marginBottom: 10,
            marginLeft: 10,
            colorBox: {
                shape: 'square', //square,circle,diamond,bean,butterfly,clover,cloverFour,cross,ellipse,heart,triangle-up,triangle-down,arrow,rocket,malteseCross,asterisk,gear
                strokeWidth: 2
            },
            colorBand: {},
            title: {
                show: false,
                text: 'Legend',
                fontSize: userProp.fontL1Size.l4,
                keepAligned: false
            },
            background: {
                color: 'transparent',
                alpha: 0
            },
            border: {
                show: true,
                size: 1,
                style: 'solid',
                radius: 0,
                color: 'transparent'
            },
            itemDisabledStyle: {
                //css properties except margin alignment and visibility properties
                color: '#d1d1d1',
                patternColor: '#ffffff',
                strikeout: false
            },
            shadow: {
                show: false,
                color: 'rgba(0,0,0,0.3)',
                x: 2,
                y: 2,
                blur: 3,
                inset: false
            },
            filter: {
                enabled: true,
                slider: {
                    tooltip: {
                        fontSize: userProp.fontL1Size.l5
                    }
                }
            },
            highlightEffect: {
                selectedSeries: null, //diagonalPattern,circlePattern,crossStripes,horizontalStripes,strokeEnlarge
                unselectedSeries: 'none' //fadeOut || blur || grayscale || saturate || none
            },
            expandable: {
                fontSize: userProp.fontL1Size.l5
            },
            size: {}
        },
        chart: {
            behaviour:{
                resetButton: {
                    enabled: false,
                    fillColor: '#3c3c3c',
                    label: {
                        marginTop: 6,
                        marginRight: 10,
                        marginBottom: 6,
                        marginLeft: 10,
                        fontSize: 14,
                        fontColor: 'white',
                        fontWeight: 500
                    },
                    border: {
                        show: true,
                        size: 1,
                        radius: 2,
                        color: 'transparent'
                    }
                }
            },
            marginTop: 10,
            marginRight: 10,
            marginBottom: 10,
            marginLeft: 10,
            fontSize: userProp.fontL2Size.l3,
            zoom: {
                enabled: false,
                type: 'x'
            },
            brush: {
                enabled: false,
                type: 'x'
            },
            effects: {
                choice: 'auto', //Choices: auto, dropShadow, chalk, freehand
                options: {
                    dropShadow: {
                        dx: '5',
                        dy: '5',
                        blur: '3',
                        color: 'rgba(0,0,0,0.6)'
                    },
                    freeHand: {
                        scale: 7
                    }
                }
            },
            plot: {
                renderer: {
                    // this option is under development
                    mode: window.rendererMode || 'svg',
                    image: {
                        source: 'auto', // auto | <any url>
                        addDimensionInSource: false, // paramas are plotWidth & plotHeight
                        exportUrl: ''
                    }
                },
                morph: {
                    enabled: false,
                    duration: 1000,
                    samplingDistance: 5
                },
                background: {
                    alpha: 1,
                    color: 'transparent',
                    gradients: {
                        show: false,
                        type: 'radial'
                    },
                    image: {}
                },
                border: {
                    show: false,
                    size: 1,
                    style: 'solid', //solid, shortdash, shortdot, shortdashdotdot, dotted, dashed, longdash, dashdot, longdashdot, longdashdotdot
                    radius: 0,
                    color: '#ececec'
                },
                animation: {
                    enabled: $ZCG.globalProp.animation.enabled,
                    easingType: $ZCG.globalProp.animation.easingType,
                    duration: $ZCG.globalProp.animation.duration
                },
                datalabels: {
                    renderOnInit: false,
                    animation: {
                        enabled: false
                    },
                    show: $ZCG.globalProp.datalabels.enabled,
                    handleOverlapping: true,
                    fontSize: userProp.datalabels.fontSize.l5,
                    boundBox: {
                        strokeColor: $ZCG.fontColor,
                        fillColor: 'white'
                    }
                },
                events: {
                    cursor: null,
                    click: $ZCG.globalProp.events.click, // On click on the plot, user function can be binded, el and event will be passed
                    mousemove: $ZCG.globalProp.events.mouseMove, // on hovering the plot, user function can be binded, el and event will be passed
                    mouseout: $ZCG.globalProp.events.mouseOut, //on mouseout the plot, user function can be binded, el and event will be passed
                    doubleclick: $ZCG.globalProp.events.doubleClick, //on mouseout the plot, user function can be binded, el and event will be passed
                    contextmenu: null, // right click event
                    init: null,
                    onload: null
                },
                eventHandler: {
                    click:{
                        dataSelction:{},
                        behaviour:{
                            selected:{},
                            unselected: '#D3D3D3'
                        }
                    }
                },
                plotoptions: {}
            },
            axes: defaultTheme.axes ? defaultTheme.axes(userProp) : {}
        },
        map: defaultTheme.map ? defaultTheme.map() : {},
        tooltip: {
            enabled: true,
            fontSize: userProp.fontL1Size.l5,
            opacity: 0.8,
            zindex: 2000,
            view: 'normal', //normal | column
            crosshair: {
                type: 'none',
                color: 'grey'
            },
            pointerEvents: 'none',
            focusOnTooltipEvent: false
        },
        noDataHandler: {
            text: 'No Data',
            fontSize: userProp.fontL1Size.l1
        },
        loader: {
            type: 'none', //zBlocks, zBars, zCircle1, zCircle2, zCrossBars, wheel, snake
            fontSize: userProp.fontL1Size.l3
        }
    };
    return defaulttheme;
};

export var getDefaultOptions = function(type, source, userProp) {
    var options;
    switch (type) {
        case 'zoom':
            options = defaultTheme.zoom(userProp);
            break;
        case 'brush':
            options = defaultTheme.brush(userProp);
            break;
        case 'scroll':
            //To implement NAVIGATOR later
            options = defaultTheme.scroll();
            break;
        case 'legend':
            options = {};
            break;

        case 'credits':
            options = defaultTheme.credits(userProp);
            break;

        case 'notes':
            options = defaultTheme.notes(userProp);
            break;

        case 'bar':
        case 'barrange':
            options = defaultTheme.bar();
            break;

        case 'line':
            options = defaultTheme.line();
            break;
        case 'area':
            options = defaultTheme.area();
            break;
        case 'arearange':
            options = defaultTheme.arearange();
            break;
        case 'scatter':
        case 'geoscatter':
            options = defaultTheme.scatter(type);
            break;
        case 'bubblepie':
        case 'geobubblepie':
            options = defaultTheme.pie(type);
            break;
        case 'pie':
            options = defaultTheme.pie(type);
            break;

        case 'funnel':
            options = defaultTheme.funnel();
            break;

        case 'web':
            options = defaultTheme.web();
            break;

        case 'pyramid':
            options = defaultTheme.pyramid();
            break;
        case 'sunburst':
            options = defaultTheme.sunburst();
            break;

        case 'dial':
            options = defaultTheme.dial(userProp);
            break;
        case 'bubble':
        case 'geobubble':
            options = defaultTheme.bubble();
            break;

        case 'packedbubble':
            options = defaultTheme.packedbubble();
            break;
        case 'bullet':
            options = defaultTheme.bullet(userProp);
            break;
        case 'gantt':
            options = defaultTheme.gantt();
            break;
        case 'treemap':
            options = defaultTheme.treemap();
            break;
        case 'heatmap':
            options = defaultTheme.heatmap();
            break;
        case 'geoheatmap':
            options = defaultTheme.geoheatmap();
            break;
        case 'dotgrid':
            options = defaultTheme.dotgrid();
            break;
        case 'waterfall':
            options = defaultTheme.waterfall();
            break;
        case 'wordcloud':
            options = defaultTheme.wordcloud();
            break;
        case 'size':
            options = defaultTheme.sizeLegend ? defaultTheme.sizeLegend(userProp) : {};
            break;
        case 'shape':
            options = defaultTheme.shapeLegend ? defaultTheme.shapeLegend(userProp) : {};
            break;
        case 'color':
            options = defaultTheme(userProp).legend;
            break;
        case 'boxplot':
            options = defaultTheme.boxplot();
            break;
        case 'sankey':
            options = defaultTheme.sankey();
            break;
        case 'metadata':
            options = {
                axes: {
                    x: [0],
                    y: [[1]],
                    tooltip: [0, 1]
                },
                columns: [
                    {
                        dataindex: 0,
                        columnname: '',
                        datatype: 'numeric'
                    },
                    {
                        dataindex: 1,
                        columnname: '',
                        datatype: 'numeric'
                    }
                ]
            };
            break;
    }
    return options;
};

export var setDefaultOptions = function(type, source, userProp, options, dontFetchDefault) {
    if (!dontFetchDefault) {
        //or To fetch default options from above
        options = getDefaultOptions(type, source, userProp);
    }
    var dest = source[type] == null ? (source[type] = {}) : source[type];
    mergeJSON(options, dest, true, true);
    source[type] = options;
    return options;
};
