// @zohocharts/zc/src/core/hierarachy-data-processor.js
//$Id$
/*
 * this is an extended utility to the data parser to process the data with supporting methods like
 * getCategories, getMin or Max of a series values etc.,
 */
//hierarachy-data-processor.js

import { chartcategory } from '../core/main-zc-init';
import { isArray, simpleClone, arrayIncludes, isUndefined, getLevel } from '../utils/zcutils';
import { d3_extent, d3_set } from './d3-constants';
import { NUMERIC_DATATYPE } from './datatype-constants';
import { DataProcessor } from './data-processor';
import { Registry } from './zc-registry';

export function HierarchyDataProcessor(data, dataSet, chartInstance) {
    this.data = data;
    this.dataset = dataSet;
    this.chartObj = chartInstance;

    this.globalChartType = null;
    this.processedseriesData = chartInstance.seriesdata;

    this.initialize();
}

HierarchyDataProcessor.prototype.flattenData = DataProcessor.prototype.flattenData;
HierarchyDataProcessor.prototype.getFlattenedSeriesData = DataProcessor.prototype.getFlattenedSeriesData;
HierarchyDataProcessor.prototype.removeFlattenData = DataProcessor.prototype.removeFlattenData;

HierarchyDataProcessor.prototype.initialize = function(state) {
    var processor = this,
        data = processor.data,
        dataset = processor.dataset,
        chartObj = processor.chartObj,
        processedseriesData = processor.processedseriesData,
        globalChartType = (processor.globalChartType = (data.seriesdata && data.seriesdata.type) || (chartObj.customVisual ? 'treemap' : null)),
        helpers = DataProcessor.helpers;

    dataset.seriesTypes = d3_set();
    processor.totalnoofdatas = 0;
    processor.isNonAxisMultiSeriesCategory = true;
    processor.isHierarchy = true;
    processor.hasHierarchyData = chartObj.preProcessor.hasHierarchyData;
    processor.seriesExtermes = [[]];
    processor.annotation = [];

    chartObj.cache.data = {};
    chartObj.cache.search = {};
    chartObj.cache.customPoints = {};
    chartObj.cache.correction = {};

    function processHierachyData(parent) {
        parent.forEach(function(d, i) {
            if (isArray(d[0])) {
                processHierachyData(d);
            } else {
                var seriesIndex = +d.level.split('_')[0],
                    currentdata = data.seriesdata.chartdata[seriesIndex],
                    yaxiscolumnorder = helpers.getParsedYAxisOrder(currentdata.yaxiscolumnorder);

                if (!currentdata.disabled) {
                    var yExtremes = d3_extent([d], function(d) {
                        return dataset.getY(d, yaxiscolumnorder);
                    });

                    processor.seriesExtermes[0].push(yExtremes);
                    processor.totalnoofdatas++;
                }
                saveAnnotation(d, chartObj, processor.annotation, seriesIndex);
            }
        });
    }

    if (processedseriesData && isArray(processedseriesData)) {
        setLevel(processedseriesData);
        processHierachyData(processedseriesData);
        processedseriesData.forEach(function(d, i) {
            var cType = helpers.getChartType(d.type || globalChartType, NUMERIC_DATATYPE);
            processor.singleSourceEvent = processor.singleSourceEvent || arrayIncludes(chartcategory.singleSourceEvent, cType);

            if(chartObj.customVisual){
                processor.singleSourceEvent = false;
            }
            dataset.seriesTypes.add(cType);
        });
        // findSubFunction(processor, processedseriesData);
    }
};

// Data adapter
export class HierarchyDataAdapter {
    constructor(data, processor) {
        if (data && isArray(data)) {
            this.data = this.convert(data, processor);
            setLevel(this.data);
        } else {
            this.data = data;
        }
    }

    convert(data, processor) {
        return dataToArray(data, processor);
    }

    getData(d) {
        return getHierarchicalData(d, this.data);
    }
}

HierarchyDataAdapter.prototype.getLevel = getLevel

function dataToArray(obj, processor) {
    return obj.map(function(d) {
        var arr;
        if (d.data && d.data[0] && isArray(d.data[0])) {
            arr = processor.hierarchical && processor.hierarchical.facetRendering ? simpleClone(d.data) : d.data; // Pivot hierarchy chart - on redraw throwing max call stack error
            var leafColors = d.leafColors;
            if (leafColors) {
                arr.forEach((d, i) => (d.color = leafColors[i % leafColors.length]));
            }
        } else if (d.data) {
            processor.hasHierarchyData = true;
            arr = dataToArray(d.data, processor);
        } else {
            arr = d;
        }

        for (var attr in d) {
            if (d.hasOwnProperty(attr) && attr !== 'data' && arr.constructor == Array || arr.constructor == Object) {
                arr[attr] = d[attr];
            }
        }

        return arr;
    });
}

function setLevel(node, level = -1) {
    node.forEach((d, i) => {
        d.level = level != -1 ? `${level}_${i}` : `${i}`;

        if (isArray(d[0])) {
            setLevel(d, d.level);
        }
    });
}
Registry.setComponent('utils.setLevel', setLevel);

function saveAnnotation(curdata, chart, arr, seriesIndex) {
    let config = chart.systemConf.notes || {},
        { dataIndex } = config;

    if (isUndefined(dataIndex)) {
        return;
    }

    let data = curdata.data || curdata;
    if (data[dataIndex]) {
        arr.push({ data: curdata, subSeriesIndex: 0, itemIndex: 0, seriesIndex });
    }
}

// Helper functions
export function getChildrenCount(dt, count) {
    if (!count) {
        count = 0;
    }

    dt.forEach(function(d) {
        if (isArray(d[0])) {
            count = getChildrenCount(d, count);
        } else {
            count++;
        }
    });
    return count;
}

function getHierarchicalData(d, seriesdata) {
    const levelArray = d ? getLevel(d).split('_') : [];
    while(levelArray.length) {
        seriesdata = seriesdata[levelArray.shift()];
    }
    return seriesdata;
}

Registry.setComponent('hierarchyDataProcessor', HierarchyDataProcessor);
Registry.setComponent('hierarchyDataAdapter', HierarchyDataAdapter);
