// @zohocharts/zc/src/core/main-constructor.js
/**
 *
 * @constructor
 */
import { Registry } from './zc-registry';
import { labelRenderer, addViewCount, Console, legend_helpers } from './main-zc-init';
import { setGlobalPropByUserdata, $ZCG } from './zcglobal';
import { $Browser } from './browser-detect';
import { d3_select, d3_namespaces, d3_event, d3_range } from './d3-constants';
import { chartThemes } from '../core/main-zc-init';
import {
    defined,
    isHidden,
    getRootBody,
    getRootNode,
    pInt,
    isArray,
    extend,
    validateValWithinLimit,
    pick,
    mergeJSON,
    isObjectPropDefined,
    getPropVal,
    getStyle,
    decodeHTML,
    validateNullVal,
    getDefsLocation,
    getDocumentIndex,
    invokeFunction,
    bbox,
    getBorderRadius,
    validateURL,
    drawBorderBox,
    getDistinct,
    destroyObjectProperties,
    each,
    arrayIncludes,
    setPropVal,
    objectValues
} from './../utils/zcutils.js';
import { DOCUMENT_OBJECT_HOLDER, BASE_CONTAINER_POSITION, POINTER, TRANSPARENT, PX, NONE, noOperation } from './zc-constants';

import { setDefaultOptions, defaultTheme } from './defaulttheme';

import { getMarkerSize } from '../utils/svg-symbols';
import { getDashArray } from '../utils/dashstyles';
import { effects, getChartEffects } from '../effects/zceffects';
import { DIV_ELEMENT } from './element-constants';
import { appendEle, trimText, addSVGtitleEle, applyFont, roundedRect, createElement } from '../utils/zcsvgutils';
import { dimension_based_intelligence } from '../utils/intelligence-utils';
// import { getLegendConfig } from '../legend/legend-helper';
import { colorscale_helpers } from '../axis/colorscale-helpers';
import { attachevents } from '../events/eventActions';
import { DataProcessor } from './data-processor';
import { legutil } from '../legend/legend-utils';
import { ZCloader } from '../loader/loader';
import { getAllSeriesTypes } from './series-parser-helpers';
import { svg_clips } from '../effects/svg-clips';
import { simpleClone } from './../utils/zcutils';
// import { getSvgTransform } from '../utils/svg-transforms';
import { getPlotBoundingRect } from '../utils/zcplotutils.js';
import { rgb2hex } from '../utils/colorutils';
import { isFacet } from '../axis/hierarchy-axis-utils';
import { Scale } from '../scale/scale-component';
// import { _legendType } from '../legend/legend-renderer-helpers';
import { updateExistingPlotComponents, updatePlotOption, checkChartTypeEquality } from '../plot/redrawTools';
import { collectExistingSeriesInfo } from '../plot/morph/morphTools';
// import { generateRanges } from '../scale-parser/legend/size/helper/size-ranges-creation';
import { getAxesCount, getAxisObject } from '../axis/axisutils';
import { createZoomScale } from '../behaviours/visual-effects/chart/tools/apply-zoom-transform-on-scale';


export var chartLayersCount;

export function VizLayout(container, userData) {
    this.container = container;
    this.userdata = userData;

    this.oldRenderer = [];
    this.dataset = {};
    this.dataObject = {};
    this.canvasarea = {};
    this.chartarea = {};
    this.legendarea = {};

    this.legend = {
        color: [
            {
                component: null
            }
        ],
        component: null
    };
    this.renderer = [];
    this.svg = null;
    this.cache = {};
    this.subrenderer = [];
    this.datalabels = {
        moduleExists: labelRenderer
    };
    this.prepareLayout();
    this.renderVisiblePointsOnly = false;
}

VizLayout.prototype.prepareLayout = function() {
    var chart = this;

    var currentChartId = addViewCount();
    chart.id = `ZC_${currentChartId}`;
    chart.idNumeric = currentChartId;

    try {
        var data = this.userdata;
        chart.isHiddenContainer = isHidden(chart.container.node());
        // chart.myFastDOM = fastdom.extend(fastdomPromised);
        chart.createHelperElements();
        chart.handleHiddenContainer();

        chart.chartWidth = chart.getChartDimension('width');
        chart.chartHeight = chart.getChartDimension('height');
        VizLayout.helpers.initializeChartHolders(chart);
        chart.setCanvasDimension();

        // get all chart types
        chart.chartTypes = DataProcessor.helpers.getAllChartTypes(data, 'object');

        chart.chartFontProp = {};
        chart.globalProp = setGlobalPropByUserdata(
            data,
            {
                width: chart.canvasarea.width,
                height: chart.canvasarea.height
            },
            chart.chartFontProp
        );

        chart.themeoptions = chart.getThemeOptions(chart.globalProp);

        // If metadata is unavailable, get default metadata
        if (!data.metadata && !chart.userdata._viewParser) {
            setDefaultOptions('metadata', data);
        }

        chart.defsLocation = getDefsLocation();

        chart.svg.append('defs');
        // chart.processUserData(chart.themeoptions);
        // chart.setChartDimension();
        // VizLayout.helpers.constructChartGroups(chart);
    } catch (err) {
        VizLayout.helpers.handleErrors(chart, err);
    }
};

var chartBasicHelpers = {
    /**
     *
     * @return {d3.selection}
     */
    getPatternSvg: function() {
        var chart = this,
            patternSvg = chart.patternSVG;

        if (!defined(patternSvg)) {
            var rootBody = getRootBody(chart.container.node());
            patternSvg = chart.patternSVG = d3_select(rootBody).select('svg.svgPattern');
        }
        return patternSvg;
    },

    /**
     *
     * @return {number}
     */
    getDocumentIndex: function() {
        var chart = this;

        if (!defined(chart.DOCUMENT_INDEX)) {
            chart.DOCUMENT_INDEX = getDocumentIndex(this.container.node());
        }
        return chart.DOCUMENT_INDEX;
    },

    /**
     *
     */
    getBaseContainer: function() {
        var chart = this,
            index = chart.getDocumentIndex();

        if (!defined(chart.BASE_HELPER_CONTAINER)) {
            chart.BASE_HELPER_CONTAINER = DOCUMENT_OBJECT_HOLDER[index].BASE_HELPER_CONTAINER;
        }
        return chart.BASE_HELPER_CONTAINER;
    },

    /**
     *
     */
    createHelperElements: function() {
        var chart = this,
            rootBody = getRootBody(chart.container.node()),
            patternSvg = d3_select(rootBody).select('svg.svgPattern'),
            isEmptyPatternSvg = patternSvg.empty();

        if (!isEmptyPatternSvg) {
            chart.patternSVG = patternSvg;
        } else {
            var rootDocument = getRootNode(rootBody);
            var parent = d3_select(rootBody);
            var baseContainer = (chart.baseContainer = appendEle(parent, 'div', [1], '.', 'class', '__baseZC_Container__'));
            baseContainer.styles({
                position: 'absolute',
                left: BASE_CONTAINER_POSITION + PX,
                top: BASE_CONTAINER_POSITION + PX
            });

            chart.patternSVG = appendEle(baseContainer, 'svg', [1], '.', 'class', 'svgPattern').attrs({
                version: '1.1',
                xmlns: d3_namespaces.svg
            });

            appendEle(chart.patternSVG, 'defs', [1]);

            /**
             *
             * settings document index and important params
             */
            if (DOCUMENT_OBJECT_HOLDER) {
                var DOCUMENT_COUNT = 0; //Need to check TODO MODULARIZE
                var documentIndex = DOCUMENT_COUNT++;
                rootDocument.ZC_DOCUMENT_INDEX = documentIndex;
                DOCUMENT_OBJECT_HOLDER[documentIndex] = {
                    BASE_HELPER_CONTAINER: chart.baseContainer,
                    EFFECTS: {},
                    ANIMATION: {},
                    CANVASPATTERN: {}
                };
            }
        }
    },

    /**
     * Handles if the container is not visible or a virtual element in document.
     * @param {Boolean} normalize to restore the cloned element
     */
    handleHiddenContainer: function(normalize) {
        var chart = this,
            container = chart.container;

        if (normalize && defined(chart.originalContainer)) {
            chart.originalContainer.node().appendChild(chart.d3container.node()); // place the cloned element under the original element.
            chart.container.remove(); // remove the cloned element.
            chart.container = chart.originalContainer; // restore the tween element.

            delete chart.isHiddenCoainer; // delete the temp variable.
            delete chart.originalContainer; // delete the temp variable.
        } else {
            var containerNode = container.node();

            if (chart.isHiddenContainer) {
                var containerClone = containerNode.cloneNode(0), // clone the parent node.
                    baseHelperContainer = chart.getBaseContainer();

                baseHelperContainer.node().appendChild(containerClone); // append to visible element.
                containerClone.style.setProperty('display', 'block', 'important'); // ensure its properly visible.

                chart.isHiddenCoainer = true;
                chart.originalContainer = container; // save the original container
                chart.container = d3_select(containerClone); // set pointer to our cloned element as main container.
            }
        }
    },

    /**
     *
     */
    updateCanvasArea: function() {
        var chart = this,
            systemConfig = chart.systemConf;

        chart.container.selectAll('div.d3container').styles({
            position: 'relative',
            width: chart.getChartDimension('width') + PX,
            height: chart.getChartDimension('height') + PX
            // "-webkit-transform" : "scale(1)",
            // "transform" : "scale(1)"
        });

        var canvas = (chart.chartParentElement = chart.container.selectAll('div.canvas-component'));
        canvas.styles({
            position: 'relative',
            width: `${chart.canvasarea.width}px`,
            height: `${chart.canvasarea.height}px`
        });

        var canvasProp = chart.container.selectAll('div.canvas-property');
        canvasProp.styles({
            position: 'absolute',
            width: `${chart.canvasarea.width}px`,
            height: `${chart.canvasarea.height}px`
        });
        chart.updateCommonBBSProp(canvasProp, null, systemConfig.canvas, 'canvas');
    },

    /**
     *
     * @param globalProp
     * @return {*}
     */
    getThemeOptions: function(globalProp) {
        var chart = this,
            allDistinctSeriesTypes = chart.distinctChartTypes = chart.distinctChartTypes || getDistinct(getAllSeriesTypes(chart.userdata)),
            theme = isObjectPropDefined(chart.userdata, 'canvas.theme', true),
            userTheme,
            defaultThemeConfig;

        defaultThemeConfig = new defaultTheme(globalProp);
        allDistinctSeriesTypes.forEach(d => setDefaultOptions(d, defaultThemeConfig.chart.plot.plotoptions, globalProp));
        var themeObj = chartThemes[theme] || (window.$ZC && $ZC[theme]);
        if (themeObj) {
            userTheme = new themeObj(globalProp);
            return mergeJSON(defaultThemeConfig, userTheme, true, true);
        }
        return defaultThemeConfig;
    },

    /**
     * it returns the dimension of the inner box (excluding padding).
     * for hidden container returns static dimension.
     *
     * @param {"width"|"height"} property
     * @returns {Number} returns width or height
     */

    getChartDimension: function(property) {
        var chart = this,
            temp = {
                width: 960,
                height: 400
            };

        if (!chart.isHiddenCoainer) {
            temp[property] = chart['_chart'+property] || getStyle(chart.container, property, true);
            chart['_chart'+property] = temp[property];
        }
        return temp[property];
    },

    /**
     *
     * @param width
     * @param height
     */

    setCanvasDimension: function(width, height) {
        var chart = this;

        if (!width && !height) {
            width = chart.getChartDimension('width');
            height = chart.getChartDimension('height');
        }

        this.canvasarea = {
            left: 0,
            top: 0,
            right: 0,
            bottom: 0,
            width: width,
            height: height
        };
    },

    /**
     *
     */
    setChartDimension: function() {
        var chart = this,
            systemConfig = chart.systemConf;
        var borderOptions = systemConfig.canvas.border;
        var borderSize_canvas = borderOptions.show
            ? validateValWithinLimit(borderOptions, 'size', 0, 10, chart.canvasarea.width, chart.canvasarea.height)
            : 0;
        chart.titlearea = {
            left: chart.canvasarea.left + borderSize_canvas,
            top: chart.canvasarea.top + borderSize_canvas,
            right: chart.canvasarea.right - borderSize_canvas,
            bottom: chart.canvasarea.bottom - borderSize_canvas,
            width: chart.canvasarea.width - 2 * borderSize_canvas,
            height: chart.canvasarea.height - 2 * borderSize_canvas
        };
        chart.chartarea = simpleClone(chart.titlearea);
        chart.plotarea = Object.create(chart.chartarea);
    },

    /**
     *
     */

    applyChartProperties: function() {
        var chart = this,
            systemConfig = chart.systemConf,
            chartHeight = chart.chartHeight;

        var title = systemConfig.canvas.title,
            subtitle = systemConfig.canvas.subtitle,
            credits = systemConfig.credits || {};
        var showTitle = title.show && dimension_based_intelligence(chart, 'title', chartHeight),
            showSubtitle = subtitle.show && dimension_based_intelligence(chart, 'subtitle', chartHeight),
            showCredits = credits.enabled === true && dimension_based_intelligence(chart, 'credits', chartHeight);
        if (showCredits) {
            var defaultCreditOptions = setDefaultOptions('credits', chart.systemConf, chart.globalProp);
            // mergeJSON(chart.systemConf.credits, defaultCreditOptions, true, true);
            chart.drawTitle(chart.systemConf.credits, 'credits');
        }
        if (showTitle) {
            title.vAlign === 'bottom' && subtitle.vAlign === 'bottom' && subtitle.show
                ? chart.drawTitle(systemConfig.canvas.subtitle, 'subtitle')
                : chart.drawTitle(systemConfig.canvas.title, 'title');
        }
        if (showSubtitle) {
            subtitle.vAlign === 'bottom' && title.vAlign === 'bottom' && title.show
                ? chart.drawTitle(systemConfig.canvas.title, 'title')
                : chart.drawTitle(systemConfig.canvas.subtitle, 'subtitle');
        }

        chart.staticPaneSVG
            .selectAll('g.plotareaBorder-property')
            .selectAll('line', 'path')
            .style('border', ''); //reset plot area border color

        chart.updatePlotArea();
    },

    updatePlotArea: function() {
        var chart = this,
            systemConfig = chart.systemConf;

        chart.chartarea = simpleClone(chart.titlearea);
        chart.plotarea = simpleClone(chart.chartarea);

        var plotOptions = systemConfig.chart;
        var mLeft = parseFloat(plotOptions.marginLeft),
            mRight = parseFloat(plotOptions.marginRight),
            mTop = parseFloat(plotOptions.marginTop),
            mBottom = parseFloat(plotOptions.marginBottom);
        // If margin values were greater than chart container width and height, reset to 0
        var mLeft_corr = mLeft + mRight > chart.plotarea.width ? 0 : mLeft,
            mRight_corr = mLeft + mRight > chart.plotarea.width ? 0 : mRight,
            mTop_corr = mTop + mBottom > chart.plotarea.height ? 0 : mTop,
            mBottom_corr = mTop + mBottom > chart.plotarea.height ? 0 : mBottom;
        chart.plotarea.left += mLeft_corr;
        chart.plotarea.right += mRight_corr;
        chart.plotarea.bottom += mBottom_corr;
        chart.plotarea.top += mTop_corr;
        chart.plotarea.width = chart.plotarea.width - mLeft_corr - mRight_corr;
        chart.plotarea.height = chart.plotarea.height - mBottom_corr - mTop_corr;

        chart.___chartarea = simpleClone(chart.plotarea);
    },

    /**
     *
     * @param data
     * @param type
     */
    drawTitle: function(data, type) {
        var chart = this,
            svg = chart.svg,
            systemConfig = chart.systemConf,
            floating = defined(data.x) && defined(data.y),
            location = window.location;

        var margin = {
            top: validateNullVal(data.marginTop, 0),
            right: validateNullVal(data.marginRight, 0),
            bottom: validateNullVal(data.marginBottom, 0),
            left: validateNullVal(data.marginLeft, 0)
        };

        var titlearea = svg.selectAll(`g.chart${type}`);
        var titleAlign = floating
            ? {
                  x: data.x + margin.left,
                  y: data.y + margin.top,
                  dy: '1',
                  textAnchor: data.hAlign === 'left' ? 'start' : data.hAlign === 'right' ? 'end' : 'middle'
              }
            : VizLayout.helpers.align(data.hAlign, data.vAlign, chart.titlearea, margin);

        // var titleBox = appendEle(titlearea, "rect", [data], ".", "class", (type+"Box"));
        var imageUrl = data.imageUrl;
        var isTextBox = !defined(imageUrl);
        var htmlEl=data.htmlEl;
        var element = isTextBox ? (data.text === '' ? NONE : 'text') : 'image';
        element=defined(htmlEl)?'html':element;

        if (element !== NONE) {
            if(!htmlEl){
                var title = appendEle(titlearea, element, [data], '.', 'class', type);
                if (defined(data.href)) {
                    var click = function() {
                        var isValidURL = validateURL(data.href);
                        if (isValidURL) {
                            var target = data.target;
                            if (target === 'newtab' || target === 'newwindow') {
                                window.open(data.href, 'popUpWindow', target === 'newwindow' ? 'status=yes' : '');
                            } else {
                                location.href = data.href;
                            }
                        } else {
                            Console.log('invalid URL for credits');
                        }
                    };
                    var mouseHandlers = {
                            click: click
                        },
                        touchHandlers = {
                            tap: click
                        };
                    attachevents(title, mouseHandlers, touchHandlers);
                }
            }
            if(htmlEl){
                title=titlearea;
                title.attrs({     
                    transform: `translate(${titleAlign.x-200},${titleAlign.y-50})`
                });
                title.html(htmlEl);
            }
            else if (isTextBox) {
                var userdataCanvasFontColor = chart.userdata.canvas ? chart.userdata.canvas.fontColor : null;

                title
                    .attrs({
                        x: titleAlign.x,
                        y: titleAlign.y,
                        dy: `${titleAlign.dy}em`,
                        'text-anchor': titleAlign.textAnchor
                    })
                    .styles({
                        cursor: defined(data.href) ? POINTER : 'auto',
                        display: 'block'
                    })
                    .text(decodeHTML(data.text));

                if (data.tooltip === true) {
                    addSVGtitleEle(title);
                }

                applyFont(title, [
                    {
                        fontColor:
                            type === 'subtitle' && !defined(data.fontColor) && !defined(userdataCanvasFontColor)
                                ? '#CCCCCC'
                                : data.fontColor || userdataCanvasFontColor
                    },
                    data,
                    systemConfig.canvas,
                    $ZCG
                ]);
            } 
            else if(imageUrl){
                title
                    .attrs({
                        x: titleAlign.x + (data.hAlign === 'right' ? -data.imageWidth : 0),
                        y: titleAlign.y + (data.vAlign === 'bottom' ? -data.imageHeight : 0),
                        width: data.imageWidth,
                        height: data.imageHeight,
                        'xlink:href': data.imageUrl
                    })
                    .styles({
                        cursor: defined(data.href) ? POINTER : 'auto',
                        display: 'block'
                    });
            }

            //validate for long text value
            var titlenode = bbox(title.node());
            var textBounds = {
                    left: floating ? data.x + margin.left : titlenode.x,
                    top: floating ? data.y + margin.top : titlenode.y,
                    width: titlenode.width,
                    height: titlenode.height
                },
                edge = {
                    left: margin.left,
                    top: margin.top,
                    right: chart.titlearea.width - margin.right,
                    bottom: chart.titlearea.height
                };
            trimText(title.node(), decodeHTML(data.text), textBounds, edge, 'bottom', '..', {
                left: 5,
                right: 5,
                top: 5,
                bottom: 5
            });

            if (!floating) {
                var nodeHeight = titlenode.height + margin.bottom + margin.top;
                if (data.vAlign === 'top') {
                    chart.titlearea.top += nodeHeight;
                    chart.titlearea.height -= nodeHeight;
                }
                if (data.vAlign === 'bottom') {
                    chart.titlearea.height -= nodeHeight;
                }
            }


            //validate for long text value
            // var titlenode = bbox(title.node());
            // let titlenode;

            // chart.myFastDOM.measure(() => {
            //     titlenode = bbox(title.node());
            // }).then(() => {
            //     var textBounds = {
            //                         left: floating ? data.x + margin.left : titlenode.x,
            //                         top: floating ? data.y + margin.top : titlenode.y,
            //                         width: titlenode.width,
            //                         height: titlenode.height
            //                     },
            //         edge = {
            //             left: margin.left,
            //             top: margin.top,
            //             right: chart.canvasarea.width - margin.right,
            //             bottom: chart.canvasarea.height
            //         };
            //     trimText(title.node(), decodeHTML(data.text), textBounds, edge, 'bottom', '..', {
            //         left: 5,
            //         right: 5,
            //         top: 5,
            //         bottom: 5
            //     });

            //     if (!floating) {
            //         var nodeHeight = titlenode.height + margin.bottom + margin.top;
            //         if (data.vAlign === 'top') {
            //             chart.titlearea.top += nodeHeight;
            //             chart.titlearea.height -= nodeHeight;
            //         }
            //         if (data.vAlign === 'bottom') {
            //             chart.titlearea.height -= nodeHeight;
            //         }
            //     }
            // })
        //     let observer;

        // let options = {
        //     root: null,
        //     rootMargin: "0px"
        // };
        // let bbox_title;
        // observer = new IntersectionObserver((entries, observer) => {
        //     entries.forEach((entry) => {
        //       const bounds = entry.boundingClientRect;
        //       // Log the `bounds` for every element
        //       bbox_title = bounds;
        //       console.log(bounds);
              
        //       var titlenode = bbox_title;
        //         var textBounds = {
        //                 left: floating ? data.x + margin.left : titlenode.x,
        //                 top: floating ? data.y + margin.top : titlenode.y,
        //                 width: titlenode.width,
        //                 height: titlenode.height
        //             },
        //             edge = {
        //                 left: margin.left,
        //                 top: margin.top,
        //                 right: chart.titlearea.width - margin.right,
        //                 bottom: chart.titlearea.height
        //             };
        //         trimText(title.node(), decodeHTML(data.text), textBounds, edge, 'bottom', '..', {
        //             left: 5,
        //             right: 5,
        //             top: 5,
        //             bottom: 5
        //         });

        //         if (!floating) {
        //             var nodeHeight = titlenode.height + margin.bottom + margin.top;
        //             if (data.vAlign === 'top') {
        //                 chart.titlearea.top += nodeHeight;
        //                 chart.titlearea.height -= nodeHeight;
        //             }
        //             if (data.vAlign === 'bottom') {
        //                 chart.titlearea.height -= nodeHeight;
        //             }
        //         }


        //     });
        //     return entries;
        //   }, options);
        //     observer.observe(title.node());
            
        }

        element!=='html'?(titlearea.selectAll(element === NONE ? 'text,image' : element === 'image' ? 'text' : 'image').style('display', NONE)):NONE;
    },

    /**
     * // BBS -->Border,Background,Shadow for canvas area, chartarea
     * @param ele
     * @param subEle
     * @param data
     * @param type
     */

    updateCommonBBSProp: function(ele, subEle, data, type) {
        var chart = this,
            systemConfig = chart.systemConf,
            border = data.border || {},
            background = data.background || {},
            shadow = data.shadow || {},
            useGradient = getPropVal(background, 'gradients.show') === true,
            borderEle;

        //background
        ele.styles({
            'background-color': background.color,
            opacity: background.alpha
        });

        if (defined(background.imageUrl)) {
            // Image url
            if (type == 'plotarea' && chart.plot.hasBgAspectRatio) {
                background.imagePosition = 'initial';
                background.imageSize = [`${systemConfig.chart.width}px`, `${systemConfig.chart.height}px`];
            }
            var styles = VizLayout.helpers.getStyles4Image(background);
            ele.styles(styles);
        } else if (useGradient) {
            // Gradient
            var styles = VizLayout.helpers.getStyles4Gradients(background);
            ele.styles(styles);
        }

        //border
        if (border.show === true && type === 'canvas') {
            borderEle = type === 'plotarea' ? subEle : ele;

            //border size validation
            var borderSize =
                type === 'canvas'
                    ? validateValWithinLimit(systemConfig.canvas.border, 'size', 0, 10, chart.canvasarea.width, chart.canvasarea.height)
                    : validateValWithinLimit(systemConfig.chart.plot.border, 'size', 0, 10, chart.plotarea.width, chart.plotarea.height);
            borderSize = pick(borderSize, 2);
            borderEle.styles({
                border: `${borderSize}px ${border.style} ${border.color}`,
                'border-radius': `${pInt(border.radius)}px`
            });
            borderEle.node().border = {
                size: borderSize,
                clr: border.color
            }; //bkpinfo
        } else {
            borderEle = type === 'plotarea' ? subEle : ele;
            if (type !== 'plotarea') {
                borderEle.style('border', '');
            }
        }

        //shadow
        if (shadow.show === true) {
            // only for canvas area
            var shadowstyle = `${shadow.x}px ${shadow.y}px ${shadow.blur}px ${shadow.color}`;
            // var shadowstyle = [shadow.x, shadow.y, shadow.blur, shadow.color].join(PX);

            if (shadow.inset === true) {
                shadowstyle += ' inset';
            }

            ele.style('box-shadow', shadowstyle);
        } else {
            ele.style('box-shadow', '');
        }
    },

    /**
     *
     */
    updatePlotareaProp: function() {
        var chart = this,
            systemConfig = chart.systemConf,
            container = chart.d3container,
            plotareaProp = chart.plotareaPropertyElement,
            plotareaBorderProp = chart.staticPaneSVG.selectAll('g.plotareaBorder-property'),
            imgLeft = chart.plotarea.left,
            imgTop = chart.plotarea.top,
            imgWidth = chart.plotarea.width,
            imgHeight = chart.plotarea.height,
            imgRight = chart.plotarea.left + chart.plotarea.width,
            imgBottom = chart.plotarea.top + chart.plotarea.height,
            plotareaBgData = systemConfig.chart.plot.background;

        if (plotareaBgData.image && plotareaBgData.image.x) {
            let scale_x = chart.axes.x.scale,
                scaleDomain_x = scale_x.domain(),
                min = invokeFunction(plotareaBgData.image.x.min, scaleDomain_x, scale_x, chart),
                max = invokeFunction(plotareaBgData.image.x.max, scaleDomain_x, scale_x, chart);
            if (defined(min)) {
                imgLeft = scale_x(min) + chart.plotarea.left;
            }
            if (defined(max)) {
                imgRight = scale_x(max) + chart.plotarea.left;
            }

            imgWidth = imgRight - imgLeft;
        }
        if (plotareaBgData.image && plotareaBgData.image.y) {
            let scale_y = chart.axes.y.groups[0].scale,
                scaleDomain_y = scale_y.domain(),
                min = invokeFunction(plotareaBgData.image.y.min, scaleDomain_y, scale_y, chart),
                max = invokeFunction(plotareaBgData.image.y.max, scaleDomain_y, scale_y, chart);
            if (defined(min)) {
                imgBottom = scale_y(min) + chart.plotarea.top;
            }
            if (defined(max)) {
                imgTop = scale_y(max) + chart.plotarea.top;
            }
            imgHeight = imgBottom - imgTop;
        }

        var styles = {
            position: 'absolute',
            left: `${imgLeft}px`,
            top: `${imgTop}px`,
            width: `${imgWidth}px`,
            height: `${imgHeight}px`,
            'pointer-events': NONE,
            'box-sizing': 'border-box'
        };
        // for adding prefix css
        styles[`${$Browser.prefix.css}box-sizing`] = 'border-box';

        chart.plotareaPropertyElement.styles(styles);
        plotareaBorderProp.attrs({
            transform: `translate(${chart.plotarea.left},${chart.plotarea.top})`
        });
        chart.updateCommonBBSProp(plotareaProp, plotareaBorderProp, systemConfig.chart.plot, 'plotarea');
    },

    /**
     *
     */

    drawPlotareaClippath: function() {
        let chart = this,
            systemConfig = chart.systemConf;
        chart.staticPaneSVG
            .selectAll('g.highlightergroup')
            .attr('filter', chart.filterEffectsID ? `url(${chart.defsLocation}#${chart.filterEffectsID})` : null);

        if (systemConfig.chart.clip === NONE) {
            return;
        }

        //If the plot is zoomed, dont consider the marker size for clipPath
        var zoomedX = isObjectPropDefined(systemConfig, 'chart.zoom.x') || defined(chart.scrolledX),
            zoomedY = isObjectPropDefined(systemConfig, 'chart.zoom.y') || defined(chart.scrolledY),
            axisrotated = chart.axes.rotated,
            ms = getMarkerSize(chart), //ms -> Marker Size
            x = (axisrotated ? zoomedY : zoomedX) ? chart.plotarea.left : -ms.left + chart.plotarea.left,
            y = (axisrotated ? zoomedX : zoomedY) ? chart.plotarea.top : -ms.top + chart.plotarea.top,
            width = chart.plotarea.width + ((axisrotated ? zoomedY : zoomedX) ? 0 : ms.left + ms.right),
            height = chart.plotarea.height + ((axisrotated ? zoomedX : zoomedY) ? 0 : ms.top + ms.bottom);

        var properties = {
                name: 'rect',
                id: chart.id,
                x: x,
                y: y,
                width: width,
                height: height
            },
            args = {
                chartObj: chart
            };

        if(chart.racing){
            properties.width += chart.plotarea.right;
        }
        var clipID = effects.get(effects.svg_clips, 'rect', properties, args),
            clipIDSelector =
                'g.seriesClippathholder, g.labelClippathholder, g.zcMapClipHighlight, g.thresholdClippathholder, g.zc-fade-overlay-clip-holder';
        clipIDSelector =
            chart.dataObject.noofGanttEnabledSeries || chart.dataObject.hasWhiskers
                ? `${clipIDSelector},g.qualitativeGroupHolder`
                : clipIDSelector;

        chart.plotClippathID = clipID;
        chart.chartParentElement.selectAll(clipIDSelector).attr('clip-path', `url(${chart.defsLocation}#${clipID})`);

        // #ZC1772, #ZC1771
        // add clip path to highlight group handles scroll
        var scroll_clip_id = `highlight_group_clip_${chart.id}`,
            plotBound = {
                x: properties.x + (chart.scrolledX || 0),
                y: properties.y + (chart.scrolledY || 0),
                width: width,
                height: height
            };
        svg_clips.rect(plotBound, scroll_clip_id, null, null, null, chart);

        let selector = 'g#zcscatterCliphighlight, g#zcareaCliphighlight, g#zcbarCliphighlight, g.zcMapClipHighlight, g#zchierarchyCliphighlight';
        selector += isFacet(chart) ? ', g.zcnonaxisCliphighlight' : ''; // ZC2879
        chart.staticPaneSVG.selectAll(selector).attr('clip-path', `url(${chart.defsLocation}#${scroll_clip_id})`);
    },

    /**
     *
     */

    getChartSeriesColor: function() {
        var chart = this,
            systemConfig = chart.systemConf,
            colors = [],
            legend_data = systemConfig.legend;

        if (legend_data.colors) {
            if (!isArray(legend_data.colors)) {
                var legendColors = [];
                for (var key in legend_data.colors) {
                    if (legend_data.colors.hasOwnProperty(key)) {
                        legendColors.push(legend_data.colors[key]);
                    }
                }
                colors = legendColors;
            } else {
                colors = legend_data.colors;
            }
        } else if (legend_data.colorPallete) {
            if (legend_data.colorPallete.type === 'monochrome') {
                colors = legutil.getMonochromeColorPallette(legend_data.colorPallete.options.monochrome);
            } else {
                colors =
                    legutil.colorpallete.get(legend_data.colorPallete.options.multicolor) || legutil.colorpallete.get('flat-ui-colors');
            }
        }

        if (legend_data.imagePallete) {
            chart.patternImageSizes = [];
            var imagePallete = legutil.imagePallete.get(legend_data.imagePallete);
            chart.imagePallete = imagePallete;

            var defaultSize = legend_data.imagePallete === 'stripes' ? null : 50;
            var tempImage = appendEle(chart.getPatternSvg(), 'image', [1], '.', 'class', 'tempImage');

            imagePallete.map(function(d, i) {
                tempImage.attr('xlink:href', d);

                var bb = bbox(tempImage.node());

                chart.patternImageSizes[i] = [defaultSize || bb.width || 3, defaultSize || bb.height || 3];
            });
            tempImage.remove();
        }
        if (legend_data.patterns) {
            chart.patternPallete = legend_data.patterns;
        } else if (legend_data.patternPallete) {
            var patternPallete = legutil.patternPallete.get(legend_data.patternPallete);
            if (!defined(patternPallete) && isArray(legend_data.patternPallete) && legend_data.patternPallete.length) {
                patternPallete = legend_data.patternPallete;
            }
            chart.patternPallete = patternPallete;
        }

        /*
         * setting seriesdata color
         */
        if (defined(colors)) {
            var colors_copy = colors.slice(0, colors.length),
                count = 0;

            chart.seriesColor = colors_copy;
            chart.mainSeriesColor = colors_copy;

            if (!colorscale_helpers.isQuantileColorScale(chart)) {
                //#ZC916
                chart.seriesdata.map(function(d, i) {
                    if (defined(d.color)) {
                        chart.seriesColor[i] = d.color;
                    } else if (!chart.seriesColor[i]) {
                        chart.seriesColor[i] = colors[count++ % colors.length]; //ZC1788
                    }
                });
            }
        }

        if (chart.dataObject.categoryHierarchy) {
            chart.dataObject.hierarchical.setGridOptions();
        }
    },

    /**
     *
     * @param updateVals - values to be updated to userdata and systemConf.
     */

    updateChartData: function(updateVals){
        var keys = Object.keys(updateVals),
            values = objectValues(updateVals);
        keys.map(function(key, idx){
            var callback = function(obj, key, val){
                if(val == null){
                    delete obj[key];
                    return;
                }
                obj[key] = val;
            }
            setPropVal(chartObj.userdata, key, values[idx], callback);
            setPropVal(chartObj.systemConf, key, values[idx], callback);
        });
    },

    /**
     *
     * @param newSystemConf
     */

    resetdata: function(newSystemConf, changeChartType) {
        var chart = this;
        chart.chartWidth = chart.getChartDimension('width');
        chart.chartHeight = chart.getChartDimension('height');
        chart.setCanvasDimension();
        // if(chart.scroll){
        //     chart.scroll.lastScroll = null
        // }
        chart.oldChartTypes = chart.chartTypes;
        chart.chartTypes = DataProcessor.helpers.getAllChartTypes(newSystemConf, 'object');
        if (changeChartType && !checkChartTypeEquality(chart.oldChartTypes, chart.chartTypes)) {
            if (chart.systemConf.chart.plot.morph.enabled) {
                chart.oldSeriesInfo = {};
                chart.oldSeriesInfo.isAxisCategory = chart.isAxisCategory;
                collectExistingSeriesInfo(chart, chart.oldChartTypes);
            }
            chart.distinctChartTypes = getDistinct(getAllSeriesTypes(chart.userdata));
            updatePlotOption(chart, newSystemConf);
            if (!(chart.seriesAddRemoveDetails && chart.seriesAddRemoveDetails.seriesAddRemovePhase)) {
                chart.svg
                    .selectAll('.seriesholder')
                    .selectAll('g')
                    .remove();
                chart.renderer = [];
                updateExistingPlotComponents(chart);
            }
        }
        if (chart.userdata.seriesdata && chart.userdata.seriesdata.chartdata) {
            newSystemConf.seriesdata.chartdata = chart.userdata.seriesdata.chartdata;
        }

        chart.processUserData(newSystemConf, true);
        chart.setChartDimension();
        chart.globalProp = setGlobalPropByUserdata(
            newSystemConf,
            {
                width: chart.canvasarea.width,
                height: chart.canvasarea.height
            },
            chart.chartFontProp
        );
        chart.events = {};
        if (chart.dataObject) {
            chart.dataObject.formattedData = {
                x: [],
                y: []
            };
        }
    },

    /**
     *
     */
    destroy: function() {
        var chart = this;

        if (chart.destroying) {
            return;
        }
        if (chart.isMiniature) {
            chart.sparkline.destroy();
            return;
        }
        invokeFunction(isObjectPropDefined(chart.systemConf, 'canvas.events.destroy', true), d3_event(), chart);

        if (chart.eventHandler) {
            clearTimeout(chart.eventHandler.mouseouttimer);
            clearTimeout(chart.eventHandler.mousemovetimer);
            if (chart.eventHandler.tooltipElement) {
                clearTimeout(chart.eventHandler.tooltipElement.timeout);
                chart.eventHandler.tooltipElement.destroy();
            }
        }
        chart.timerObj.forEach(d => {
            clearTimeout(d);
            clearInterval(d);
        });

        // remove the container reference
        chart.container.datum(null);
        chart.container = chart.baseContainer = chart.BASE_HELPER_CONTAINER = chart.patternSVG = chart.originalContainer = null;

        if (chart.dataObject.isAxisCategory || chart.dataObject.isPolarAxisCategory) {
            // remove axis
            destroyObjectProperties(chart.axes.x);
            chart.axes.y.groups.forEach(yAxis => destroyObjectProperties(yAxis));
        }

        // remove label array #ZC3245
        chart.renderer.forEach(renderer => destroyObjectProperties(renderer.commonRendererProp().params.labelArray));
        //remove label queue, label parser @memory leak issue
        chart.renderer.forEach(renderer => {
            destroyObjectProperties(renderer.commonRendererProp().params.labelQueue, chart)
            destroyObjectProperties(renderer.commonRendererProp().params.labelParser, chart)
        });
        // remove all detached element references before removing renderer
        chart.renderer.forEach(renderer => renderer.destroy && renderer.destroy()); // remove detached elements from memory
        // remove all renderer
        chart.renderer.forEach(renderer => destroyObjectProperties(renderer, chart));

        // #ZC2766 destroy the behavioural events
        if (chart.behaviouralEvents) {
            destroyObjectProperties(chart.behaviouralEvents.eventManager.brushResetButton, chart);
            each(chart.behaviouralEvents.eventManager.behaviours, behaviour => {
                destroyObjectProperties(behaviour);
            });
            destroyObjectProperties(chart.behaviouralEvents, chart);
        }

        destroyObjectProperties(chart.eventHandler, chart);
        destroyObjectProperties(chart.dataObject, chart);
        destroyObjectProperties(chart.cache);
        destroyObjectProperties(chart);

        chart.destroying = true;
        d3_event(null);
    },

    /**
     *
     */

    redraw: async function(changeChartType) {
        var chart = this,
            systemConfig = chart.systemConf;

        if (chart.isMiniature) {
            chart.sparkline.redraw();
            return;
        }

        chart.updating = true;
        chart.phase = 'redraw';

        if(chart.racing){
            await chart.racing.handlers.onRedraw(chart);
        }

        chart.viewPortExtent = null;

        chart._chartwidth = null;
        chart._chartheight = null;

        chart.svg.styles({
            width: chart.getChartDimension('width') + PX,
            height: chart.getChartDimension('height') + PX
        });
        chart.staticPaneSVG.styles({
            width: chart.getChartDimension('width') + PX,
            height: chart.getChartDimension('height') + PX
        });
        chart.resetdata(systemConfig, changeChartType);
        VizLayout.helpers.constructChartGroups(chart);
        chart.render(true);
        chart.phase = null;
    },

    drawPlotBorder: function() {
        var chart = this,
            borderConfig = chart.systemConf.chart.plot.border,
            borderElement = chart.staticPaneSVG.selectAll('g.plotareaBorder-property');

        if (defined(borderConfig) && borderConfig.show === true) {
            var borderEle = borderElement.selectAll('path.plotBorderPath'),
                borderSize = validateValWithinLimit(borderConfig, 'size', 0, 10, chart.plotarea.width, chart.plotarea.height),
                tickArea = pick(chart.tickarea, { left: 0, right: 0, top: 0, bottom: 0 }),
                bleft = -tickArea.left || 0.5,
                bTop = -tickArea.top || 0.5,
                bWidth = chart.plotarea.width + tickArea.right + tickArea.left,
                bHeight = chart.plotarea.height + tickArea.bottom + tickArea.top,
                radius = getBorderRadius(borderConfig.radius),
                rectAttr = roundedRect(bleft, bTop, bWidth, bHeight, radius, false, false);

            borderSize = borderSize == null || borderSize === '' ? 2 : borderSize;

            borderEle
                .styles({
                    'stroke-width': borderSize,
                    stroke: borderConfig.color,
                    'stroke-dasharray': getDashArray(borderConfig.style, borderSize),
                    fill: NONE,
                    'pointer-events': NONE
                })
                .attrs({
                    d: rectAttr
                });
            borderElement.selectAll('line').style('stroke', TRANSPARENT);
        }
    },

    /**
     *
     * @param data
     * @param ele
     */
    handleNoData: function(data, ele) {
        this.clearChart();
        var chart = this,
            systemConfig = chart.systemConf,
            nodatawrap = pick(ele, chart.wrap.selectAll('g.nodatahandler'));

        if (defined(data.htmlEl)) {
            var customHandler = appendEle(chart.container.selectAll('div.canvas-component'), 'div', [1], '.', 'class', 'nodatahandler');
            customHandler.style('position', 'absolute').html(data.htmlEl);

            var node = customHandler.node();
            customHandler.styles({
                left: `${pick(data.x, (chart.chartWidth - node.offsetWidth) / 2)}px`,
                top: `${pick(data.y, (chart.chartHeight - node.offsetHeight) / 2)}px`
            });
        } else if (defined(data.imageUrl)) {
            var image = appendEle(nodatawrap, 'image', [1]);
            image.attrs({
                x: data.x || chart.chartarea.width / 2 - 50,
                y: data.y || chart.chartarea.height / 2 - 50,
                width: 100,
                height: 100,
                'xlink:href': data.imageUrl
            });
        } else if (defined(data.text)) {
            var text = appendEle(nodatawrap, 'text', [data], '.', 'class', 'nodata');
            text.attrs({
                x: data.x || chart.chartarea.width / 2,
                y: data.y || chart.chartarea.height / 2,
                'text-anchor': 'middle'
            }).text(decodeHTML(data.text || 'No Data'));
            applyFont(text, [data, systemConfig.canvas, $ZCG]);
        }
    },

    /**
     *
     * @param data
     * @param ele
     */

    clearNoData: function(data, ele) {
        var chart = this,
            systemConfig = chart.systemConf,
            nodatawrap = pick(ele, chart.wrap.selectAll('g.nodatahandler'));

        if (defined(data.htmlEl)) {
            var customHandler = appendEle(chart.container.selectAll('div.canvas-component'), 'div', [], '.', 'class', 'nodatahandler');
        } else if (defined(data.imageUrl)) {
            var image = appendEle(nodatawrap, 'image', []);
        } else if (defined(data.text)) {
            var text = appendEle(nodatawrap, 'text', [], '.', 'class', 'nodata');
        }
    },

    /**
     *
     * Remove chart elements
     *
     */

    clearChart: function(data, ele, keepChartGroup) {
        var chart = this;

        if (!keepChartGroup) {
            appendEle(chart.wrap, 'g', [], '.', 'class', 'chartgroup');
        }
        appendEle(chart.staticPaneSVG, 'g', [], '.', 'class', 'plotareaBorder-property');
        appendEle(chart.staticPaneSVG, 'g', [], '.', 'class', 'labelClippathholder');
        appendEle(chart.staticPaneSVG, 'g', [], '.', 'class', 'qualitativeGroupHolder');
        appendEle(chart.container, 'div', [], '.', 'class', 'legendgroup');
        appendEle(chart.staticPaneSVG, 'g', [], '.', 'class', 'legendarea');
        appendEle(chart.container, 'div', [], '.', 'class', 'zcnotes');
        //canvas
        appendEle(chart.container, 'canvas', [], '.', 'class', 'zc-canvas');
        appendEle(chart.container, 'canvas', [], '.', 'class', 'zc-labelholder-canvas');
        if (chart.eventHandler && (chart.isAxisCategory || chart.dataObject.isPolarAxisCategory || chart.isGeoMap)) {
            // Remove event listenener for chart with nodata on redraw
            // drawBorderBox(chart.eventHandler.trackerElement, { width: 0, height: 0 });
            chart.positionEventTrackerElement(null, {});
        }
    },

    /**
     *
     * @param container
     */

    showLoading: function(container) {
        var chart = this,
            systemConfig = chart.systemConf;

        var data = systemConfig.loader;
        if (data.enabled === true && data.type !== NONE) {
            var loader = appendEle(this.container.selectAll('div.d3container'), 'div', [1], '.', 'class', 'loading');
            loader.style('position', 'absolute').html(VizLayout.helpers.getLoader(chart, data));
            applyFont(loader, [data, systemConfig.chart, systemConfig.canvas, $ZCG], true);
            var node = loader.node();
            loader.styles({
                left: `${chart.chartarea.width / 2 - node.offsetWidth / 2}px`,
                top: `${chart.chartarea.height / 2 - node.offsetHeight / 2}px`
            });
        }
    },

    hideLoading: function() {
        this.container.selectAll('div.loading').remove();
    },

    //! Commented unused code
    // /**
    //  *
    //  */
    // fixLegendPosition: function() {
    //     // Not handled multi legends
    //     var chart = this,
    //         legendArea = chart.legendarea,
    //         legendConfig = chart.systemConf.legend;

    //     if (legendConfig.enabled === true) {
    //         var pos = getLegendPosition(chart.plotarea, chart, false),
    //             isContinuousLegend = legend_helpers.isContinousLegend(chart);

    //         if (legendArea.adjust === 'y') {
    //             if (isContinuousLegend) {
    //                 var translate = getSvgTransform(legendArea.element.node()).translate;
    //                 legendArea.element.attr('transform', `translate(${translate[0]},${pos[1]})`);
    //             } else {
    //                 legendArea.element.style('top', pos[1] + PX);
    //             }
    //         }
    //     }
    // },

    createEventGroups: function(update) {
        var chart = this,
            systemConfig = chart.systemConf,
            plotInitStyles = { opacity: 0, 'pointer-events': NONE },
            initStyles = { ...plotInitStyles, fill: NONE };

        var eventgroup = (chart.eventWrap = chart.staticPaneSVG.selectAll('g.eventgroup').styles({
            'pointer-events': 'visible'
        }));

        // To apply chart effects to highlighters, seperate group has been created. to avoid effect on reset text button
        let highlighterGroup = appendEle(eventgroup, 'g', [1], '.', 'class', 'highlightergroup');

        var areaClipPath = highlighterGroup.append('g').attr('id', 'zcareaCliphighlight'),
            areahighlight = areaClipPath.append('g').attr('id', 'zcareahighlight');

        var barClipPath = highlighterGroup.append('g').attr('id', 'zcbarCliphighlight'),
            barhighlight = barClipPath.append('g').attr('id', 'zcbarhighlight');

        if (systemConfig.tooltip.crosshair) {
            var crosshairEle = appendEle(highlighterGroup, 'g', [1], '.', 'class', 'crosshair');
            crosshairEle.style('visibility', 'hidden');
            appendEle(crosshairEle, 'line', [1], '.', 'class', 'x-crosshair');
            appendEle(crosshairEle, 'line', [1], '.', 'class', 'y-crosshair');
        }

        var scatterClipPath = highlighterGroup.append('g').attr('id', 'zcscatterCliphighlight'),
            scatterhighlight = scatterClipPath.append('g').attr('id', 'zcscatterhighlight'),
            selectionhighlight = scatterClipPath.append('g').attr('id', 'zcselectionhighlight');

        let plotHighlight = createElement('g', '.zcplothighlight', [1], highlighterGroup);
        plotHighlight.styles(plotInitStyles);

        let nonAxisClipPath = createElement('g', '.zcnonaxisCliphighlight', [1], highlighterGroup),
            nonAxisHighlight = createElement('g', '.zcnonaxishighlight', [1], nonAxisClipPath);
        nonAxisHighlight.styles(initStyles);
        //if (chart.dataObject.isHierarchy) {
        var hierarchyClipHighlight = highlighterGroup.append('g').attr('id', 'zchierarchyCliphighlight'),
            hierarchyHighlight = hierarchyClipHighlight.append('g').attr('id', 'zchierarchyhighlight');
        hierarchyHighlight.styles(initStyles);
        //}
    },

    //? Does this function needed in custom bundling without events
    positionEventTrackerElement(tracker, plotArea = getPlotBoundingRect(this, true)) {
        if (!this.eventHandler) { return };
        let chart = this,
            group = chart.eventHandler.trackerGroup,
            plotarea = chart.plotarea,
            isWidthScroll = chart.scrolledX,
            isHeightScroll = chart.scrolledY,
            hasDefaultScroll = isWidthScroll || isHeightScroll;

        // group should always have the plot area transform
        group.attr('transform', `translate(${plotarea.left || 0} ${plotarea.top || 0})`);

        tracker = pick(tracker, chart.eventHandler.trackerElement);

        // transform the parent
        // for div scroll apply the transform to the rect
        // # ZC2036
        if (hasDefaultScroll) {
            tracker.attr('transform', `translate(${chart.scrolledX || 0} ${chart.scrolledY || 0})`);
        } else {
            tracker.attr('transform', `translate(${plotArea.left || 0} ${plotArea.top || 0})`);
        }

        drawBorderBox(tracker, plotArea);
    },

    /**
     *
     * @param {"x"|"y"|"xy"} type
     * @param {"zoom"|"brush"} module
     * @private
     */
    __enableZoomBrush: function(type, module) {
        let chart = this,
            key = `PLOT_${module.toUpperCase()}`,
            behaviouralEvents = chart.behaviouralEvents,
            behaviours = behaviouralEvents.eventManager.behaviours;

        const enable = () => {
            // merge the default options.
            chart.systemConf.chart[module].enabled = true;
            chart.systemConf.chart[module].type = type;
            behaviouralEvents.initiateBehaviour(key);
        };

        // TODO disable
        // const disable = () => {
        //     chart.systemConf.chart[module].enabled = false;
        //     chart[key].destroy();
        //     delete chart[key];
        // };

        if (!behaviours[key]) {
            enable();
        } else if (behaviours[key] && behaviours[key].instance.options.type !== type) {
            enable();
        } else if (type === null) {
            // TODO disable
        }
    },

    __enableZoom: function(type) {
        this.__enableZoomBrush(type, 'zoom');
        var axisrotated = this.axes.rotated,
            _zoomScale = function(chart, type, index){
                var axis = getAxisObject(chart, type, index);
                createZoomScale(axis, chart);
            };
        if (type === 'xy') {
            var xCount = getAxesCount(this, 'x', axisrotated ? 'left' : 'bottom'),
                yCount = getAxesCount(this, 'y', !axisrotated ? 'left' : 'bottom'), i;
            for (i = 0; i < xCount; i++) {
                _zoomScale(this, 'x', i);
            }
            for (i = 0; i < yCount; i++) {
                _zoomScale(this, 'y', i);
            }
        } else if(defined(type)){
            var axesCount = getAxesCount(this, type, (type === 'x' && axisrotated) || (type === 'y' && !axisrotated) ? 'left' : 'bottom');
            for (var i = 0; i < axesCount; i++) {
                _zoomScale(this, type, i);
            }
        } 
    },

    __enableBrush: function(type) {
        this.__enableZoomBrush(type, 'brush');
    },

    /**
     *
     * @param {number} step
     */
    zoomStep(step) {
        let type = this.isGeoMap ? 'xy' : getPropVal(this.systemConf, 'chart.zoom.type') || 'x';
        this.__enableZoom(type);
        this.behaviouralEvents.eventManager.behaviours.PLOT_ZOOM.instance.step(step);
    },

    parseScale: function() {
        let chartObj = this,
            dataSplicer = chartObj.dataSplicer,
            metadata = dataSplicer.getEncode(),
            rangeCreation = ['z', 'shape'],
            SUPPORTED = chartObj.dataObject.hasZaxis ? ['z'] : [],
            encodes = {},
            scales = [],
            getBaseConfig = encode => {
                if (encode === 'size' || encode === 'color' || encode === 'shape') {
                    return chartObj.systemConf.legend;
                }
            },
            getConfig = (_, key) => {
                let encode = _ === 'z' ? 'size' : _ === 'clr' ? 'color' : _,
                    baseConfig = getBaseConfig(encode);

                setDefaultOptions(encode, baseConfig, chartObj.globalProp);
                return baseConfig[encode][key] || baseConfig[encode];
            };

        if (defined(chartObj.dataset.getColumnIdx('shape'))) {
            SUPPORTED.push('shape');
        }

        SUPPORTED.forEach(_ => {
            if (metadata[_]) {
                encodes[_] = d3_range(metadata[_].length);
            }
        });

        for (var _ in encodes) {
            encodes[_].forEach(key => {
                let _parser = Registry.getComponent(`${_ === 'z' ? 'size' : _}.parser`);
                if (_parser != noOperation) {
                    let config = getConfig(_, key),
                        { ranges, stops, minRange, maxRange, base, nullAs } = config,
                        unknown = undefined;

                    if (ranges && !isArray(ranges[0])) {
                        ranges = ranges.map(r => [r]);
                        config.ranges = ranges;
                    }

                    if (ranges && ranges._dummy) {
                        ranges = null;
                        // config.ranges = null;
                    }

                    let parser = _parser(_, key, chartObj, { ...config, ranges }),
                        scaleComponent = new Scale(_, key, { ranges, unknown, stops, minRange, maxRange, base, nullAs, ...parser }, chartObj); // TODO: check this
                    scales.push(scaleComponent);
                }
            });
        }

        scales.forEach(_ => _.getScale());
        scales.forEach((_, i) => {
            let { encode, key } = _;
            if (arrayIncludes(rangeCreation, encode)) {
                let scale = _.getScale(),
                    { schema } = _,
                    config = legend_helpers.getLegendConfig(encode, key, chartObj),
                    _legendType = Registry.getComponent('legend.helpers.legendGroupType'),
                    view = _legendType(encode, { config, dataType: schema.data.dataType }),
                    generateRanges = Registry.getComponent('legend.size.helpers.generateRanges'),
                    newRanges = generateRanges(config, schema, scale, view);

                if (newRanges) {
                    config.ranges = newRanges;
                    schema.domain.ranges = newRanges;

                    let _parser = Registry.getComponent(`${encode === 'z' ? 'size' : encode}.parser`);
                    let { ranges, stops, minRange, maxRange, base, nullAs } = config,
                        parser = _parser(encode, key, chartObj, config),
                        scaleComponent = new Scale(encode, key, { ranges, stops, minRange, maxRange, base, nullAs, ...parser }, chartObj);
                    scaleComponent.create();
                    scales[i] = scaleComponent;
                }
            }
        });
        chartObj.scales = scales;
    },

    handleLegendEvents() {
        let chart = this,
            _legendEvents = Registry.getComponent('LegendEvents');

        if (!_legendEvents._empty) {
            if (chart.legendEvents) {
                chart.legendEvents.update();
            } else {
                chart.legendEvents = new _legendEvents(chart);
            }
        }

        if (isFacet(chart)) {
            let disabledOpacity = pick(isObjectPropDefined(chart.systemConf, 'legend.selectAllBox.disabledOpacity', true), 0.3);
            legend_helpers.fadeChart(null, chart, disabledOpacity);
        }
    },

    _isAxisCategory() {
        let { dataObject, preProcessor } = this,
            { hierarchical } = dataObject,
            plotOptions = this.systemConf.chart.plot.plotoptions[this.distinctChartTypes && this.distinctChartTypes[0]];//this plotOptions are used to check if axis is enabled in treemap or not
        //since combinational chart is not possible using treemap - directly taken this.distinctChartTypes[0]
        return (
            dataObject.isAxisCategory ||
            (hierarchical && hierarchical.facetRendering) ||
            (preProcessor && preProcessor.hierarchical && preProcessor.hierarchical.facetRendering) ||
            (plotOptions && plotOptions.axis && plotOptions.axis.enabled)
        );
    }
};

export let createHelperElements = chartBasicHelpers.createHelperElements;

extend(VizLayout.prototype, chartBasicHelpers);

/************************** Helpers ****************/

VizLayout.helpers = {
    initializeChartHolders: function(chart) {
        var container = chart.container,
            width_style,
            height_style;

        var d3container = container.selectAll('div.d3container');
        if (!d3container.empty()) {
            // If user going to create a new chart in the same container again, destroy the previous chart object
            var prevChartObj = d3container.datum();
            if (prevChartObj && prevChartObj.destroy) {
                prevChartObj.destroy();
            }

            d3container.remove();
        }

        width_style = chart.getChartDimension('width');
        height_style = chart.getChartDimension('height');

        d3container = chart.d3container = appendEle(container, 'div', [chart], '.', 'class', 'd3container');
       

        //CANVASPROP - canvas area properties-border,background,shadow
        appendEle(d3container, 'div', [1], '.', 'class', 'canvas-property');
        //PLOTAREA PROP - plotarea background prop
        chart.plotareaPropertyElement = appendEle(d3container, 'div', [1], '.', 'class', 'plotarea-property');

        // Overview container for scroll
        appendEle(chart.d3container, 'div', [1], '.', 'class', 'zcoverview');

        //main canvas area - svg
        var canvas = appendEle(d3container, 'div', [1], '.', 'class', 'canvas-component');
        canvas.style('-webkit-tap-highlight-color', TRANSPARENT);

        var tooltipElement = appendEle(chart.getBaseContainer(), 'div', [1], '#', 'id', `zctooltip${chart.idNumeric}`);
        tooltipElement.node().isTooltipContainer = true;

        chart.tooltip = {
            element: tooltipElement
        };
        

        
        chart.svg = appendEle(canvas, 'svg', [1], '.', 'class', 'svgarea');
        chart.staticPaneSVG = appendEle(canvas, 'svg', [chart], '.', 'class', 'static').style('pointer-events', NONE);

        appendEle(canvas, DIV_ELEMENT, [1], '.', 'class', 'legendarea');
        // canvas.append(DIV_ELEMENT).attr('class', 'legendarea');
        // chart.notes = canvas.append('div').attr('class', 'zcnotes');
        chart.notes = appendEle(canvas, DIV_ELEMENT, [1], '.', 'class', 'zcnotes');


        

        var styles = {
            width: `${width_style}px`,
            height: `${height_style}px`,
            position: 'relative',
            left: '0px',
            top: '0px',
            'box-sizing': 'border-box',
            '-moz-user-select': NONE,
            '-webkit-user-select': NONE,
            '-ms-user-select': NONE,
            'user-select': NONE
        };
        // for adding prefix css
        styles[`${$Browser.prefix.css}box-sizing`] = 'border-box';
        d3container.styles(styles);


        var SVGposition = VizLayout.helpers.positionSVG(chart, width_style, height_style);
        chart.svg.styles(SVGposition)

        chart.staticPaneSVG.styles(SVGposition)

        chart.tooltip.element.styles({
            position: 'fixed',
            display: 'inline-block',
            visibility: 'hidden',
            'white-space': 'nowrap',
            'background-color': 'white',
            border: '1px solid',
            '-webkit-tap-highlight-color': TRANSPARENT
        });
    },

    removeChartHolders: function(chart) {
        var container = chart.container;
        container
            .select('div.d3container') //No I18N
            .remove();

        container
            .select('div.zc-staticPane') //No I18N
            .remove();
        container
            .select('div.zc-chartPane') //No I18N
            .remove();
    },

    /**
     *
     * @param hAlign
     * @param vAlign
     * @param box
     * @param margin
     * @returns {{}}
     */

    align: function(hAlign, vAlign, box, margin) {
        var attr = {};
        //hAlign
        if (hAlign === 'center') {
            attr.x = box.width / 2 + margin.left;
            attr.textAnchor = 'middle';
        } else if (hAlign === 'left') {
            attr.x = box.left + margin.left;
            attr.textAnchor = 'start';
        } else if (hAlign === 'right') {
            attr.x = box.width + box.left - margin.right;
            attr.textAnchor = 'end';
        }
        //vAlign
        if (vAlign === 'top') {
            attr.y = box.top + margin.top;
            attr.dy = 1;
        } else if (vAlign === 'bottom') {
            attr.y = box.height + box.top - margin.bottom;
            attr.dy = -0.21;
        }
        return attr;
    },

    getLoader: function(chart, data) {
        var htmlEl = data.htmlEl,
            imageUrl = data.imageUrl,
            type = data.type,
            text = data.text,
            colors = data.colors;
        if (htmlEl) {
            return htmlEl;
        } else if (imageUrl) {
            return `<img src=${imageUrl} height=${data.imageHeight} width=${data.imageWidth}>`;
        } else {
            var html = '',
                ele = null;
            if (type) {
                var loaderDiv = chart.container.selectAll('div.loading');
                colors = colors && colors.length ? colors : null;
                ele = ZCloader(loaderDiv, {
                    type: type,
                    colors: colors
                });

                var XMLS = new XMLSerializer();
                html = XMLS.serializeToString(ele); // Converting DOM node into a string
            }
            if (text) {
                html += `<p style=position:relative;display:inline-block>${text}</p>`;
            }
            return html;
        }
    },

    constructChartGroups: function(chart) {
        var svg = chart.svg,
            systemConfig = chart.systemConf,
            chartHeight = chart.chartHeight,
            chartWidth = chart.chartWidth;
        //Title area
        var istitle = systemConfig.canvas.title.show === true && dimension_based_intelligence(chart, 'title', chartHeight);
        appendEle(svg, 'g', groupdata(istitle), '.', 'class', 'charttitle');

        //Subtitle area
        var issubtitle = systemConfig.canvas.subtitle.show === true && dimension_based_intelligence(chart, 'subtitle', chartHeight);
        appendEle(svg, 'g', groupdata(issubtitle), '.', 'class', 'chartsubtitle');
        //Credit Area
        var iscredits =
            systemConfig.credits && systemConfig.credits.enabled === true && dimension_based_intelligence(chart, 'credits', chartHeight);
        appendEle(svg, 'g', groupdata(iscredits), '.', 'class', 'chartcredits');

        var wrap;

        if (chart.isGeoMap && chart.geo.isCanvasMode && chart.dataObject.hasPointChart) {
            wrap = chart.wrap = createElement('g', '.xychart', [1], chart.staticPaneSVG, ':first-child');
        } else {
            wrap = chart.wrap = appendEle(svg, 'g', [1], '.', 'class', 'xychart');
        }

        if (chart.nochartdata) {
            appendEle(wrap, 'g', [1], '.', 'class', 'nodatahandler');
        } else {
            appendEle(wrap, 'g', [0], '.', 'class', 'nodatahandler');
            var chartgroup = appendEle(wrap, 'g', [chart.seriesdata], '.', 'class', 'chartgroup');

            var isxaxis = isObjectPropDefined(systemConfig, 'chart.axes.xaxis.show', true) === true;
            //xaxisgroup
            appendEle(chartgroup, 'g', groupdata(isxaxis), '.', 'class', 'x-axis');
            //yaxisgroup
            appendEle(chartgroup, 'g', [chart.seriesdata], '.', 'class', 'y-axis');

            // Underlay threshold - So that threshold range will render first and then plot will be rendered
            appendEle(chartgroup, 'g', [chart.seriesdata], '.', 'class', 'threshold');

            var seriesClipgroup = appendEle(chartgroup, 'g', [chart.seriesdata], '.', 'class', 'seriesClippathholder'); //#ZC1502

            //series group
            var seriesgroup = chart.seriesholder = appendEle(seriesClipgroup, 'g', [chart.seriesdata], '.', 'class', 'seriesholder');
            chart.filterEffectsID = getChartEffects(chart);
            let effectValue = chart.filterEffectsID ? `url(${chart.defsLocation}#${chart.filterEffectsID})` : null;
            seriesgroup.attr('filter', effectValue);

            var labelClipgroup = appendEle(chart.staticPaneSVG, 'g', [chart.seriesdata], '.', 'class', 'labelClippathholder'); //#ZC1502

            var onlyBarChart = false;
            if (chart.dataObject) {
                onlyBarChart =
                    (chart.dataObject && chart.dataObject.noofBarEnabledSeries && !chart.dataObject.isCombinationalChart) ||
                    chart.dataObject.isHierarchy ||
                    chart.dataObject.isFlowChartCategory; // For bar charts, the element should be before the highlighting element
            }

            if (!onlyBarChart) {
                appendEle(labelClipgroup, 'g', [chart.seriesdata], '.', 'class', 'labelholder');
            }

            //thgroup
            let thresholdClipG = appendEle(chart.staticPaneSVG, 'g', [chart.seriesdata], '.', 'class', 'thresholdClippathholder');
            appendEle(thresholdClipG, 'g', [chart.seriesdata], '.', 'class', 'threshold');

            // Qualitative group (without clip path)
            var qualitativeGroup = appendEle(chart.staticPaneSVG, 'g', [1], '.', 'class', 'qualitativeGroupHolder');
            appendEle(qualitativeGroup, 'g', [1], '.', 'class', 'qualitativeGroup');

            //plotarea border properties,as border has to be drawn over axis tickmarks // also below the event rect
            var plotBorder = appendEle(chart.staticPaneSVG, 'g', [1], '.', 'class', 'plotareaBorder-property');
            var isplotBorder = systemConfig.chart.plot.border.show === true;
            appendEle(plotBorder, 'path', groupdata(isplotBorder), '.', 'class', 'plotBorderPath');

            //eventgroup
            appendEle(chart.staticPaneSVG, 'g', [1], '.', 'class', 'eventgroup');

            if (onlyBarChart) {
                //labelgroup
                labelClipgroup.raise();
                var labelgroup = appendEle(labelClipgroup, 'g', [chart.seriesdata], '.', 'class', 'labelholder');
                labelgroup.style('pointer-events', NONE);
            }

            // Dummy rect added for filter - ISSUE : http://www.inkscapeforum.com/viewtopic.php?t=18611
            var dummyRectangleForFilters = appendEle(seriesgroup, 'rect', [1], '.', 'class', 'dummy-rect-filter');
            dummyRectangleForFilters.style('pointer-events', NONE);

            VizLayout.helpers.createLegendArea(chart);
        }

        function groupdata(show) {
            return show ? [1] : [];
        }
    },

    handleErrors: function(chart, err) {
        // [IE Fix $err.stack & Console.error]
        err = err.stack ? err.stack : err;
        Console.error ? Console.error(err) : Console.log(err);

        // console.log("Oops! Sorry something went wrong to render the chart...")
        // console.log("Error message : " + err.message);
        // Console.log(err.stack);

        chart.breakOnErr = true;
        if (isObjectPropDefined(chart.systemConf, 'canvas.events')) {
            invokeFunction(chart.systemConf.canvas.events.onerror, d3_event(), chart, err);
        }
    },

    positionSVG: function(chart, width, height) {
        return {
            width: width + PX,
            height: height + PX,
            position: 'absolute',
            top: 0,
            left: 0
        };
    },

    //? Needed when legend component is added during custom bundle. Unwanted import of legend_helpers
    createLegendArea: function(chart) {
        var isContinousLegend = legend_helpers.isContinousLegend && legend_helpers.isContinousLegend(chart),
            elementContinuous = createElement('g', '.legendarea', [1], chart.staticPaneSVG),
            elementDiscrete = chart.container.selectAll(`${DIV_ELEMENT}.legendarea`);

        // applyFont(elementContinuous, [chart.systemConf.canvas, $ZCG]);
        // applyFont(elementDiscrete, [chart.systemConf.canvas, $ZCG], true);

        if (isContinousLegend) {
            var dataObject = chart.dataObject,
                onlyBarChart =
                    dataObject && ((dataObject.noofBarEnabledSeries && !dataObject.isCombinationalChart) || dataObject.isHierarchy);

            if (onlyBarChart) {
                elementContinuous.raise();
            }

            elementDiscrete.style('display', NONE);
            chart.legendarea.element = elementContinuous;
        } else {
            elementContinuous.style('display', NONE);
            chart.legendarea.element = elementDiscrete;
        }
    },
    getChartId: function(chart) {
        return chart.id;
    },
    setChartId: function(chart) {
        var currentId = chart.id;
    },

    getStyles4Gradients: function(options) {
        var gradients = options.gradients || {};

        if ($Browser.IE === 1 && $Browser.VERSION.indexOf('9.') === 0) {
            var stopColor = gradients.stopColor ? gradients.stopColor.map(rgb2hex) : ['black', 'white']; // IE needs hex color code

            if (gradients.type === 'linear') {
                var filter = `progid:DXImageTransform.Microsoft.gradient(startColorstr=${stopColor[0]}, endColorstr=${stopColor[1]})`;

                return { filter };
            } else {
                var filter = 'progid:DXImageTransform.Microsoft.Alpha(opacity=0, finishopacity=100, style=2)';

                return {
                    filter,
                    'background-color': stopColor[1]
                };
            }
        } else {
            var prefix = $Browser.WEBKIT ? '-webkit-' : $Browser.FIREFOX ? '-moz-' : $Browser.IE ? '-ms-' : '',
                gradient,
                gradientColors = '';

            if (!defined(gradients.useCSS)) {
                if (defined(gradients.stopColor)) {
                    gradients.stopColor.map(function(d, i) {
                        var offset = defined(gradients.stopOffset) && defined(gradients.stopOffset[i]) ? ` ${gradients.stopOffset[i]}` : '';
                        gradientColors = gradientColors + (i === 0 ? `${d} ${offset}` : `,${d}${offset}`);
                    });
                }
                if (gradients.type === 'radial') {
                    gradient = `radial-gradient(${gradientColors})`;
                } else if (gradients.type === 'linear') {
                    gradient = `linear-gradient(${gradientColors})`;
                }
            } else {
                gradient = gradients.useCSS;
            }

            return {
                'background-image': prefix + gradient
            };
        }
    },

    getStyles4Image: function(options) {
        return {
            'background-image': `url(${options.imageUrl})`,
            'background-repeat': options.imageRepeat,
            'background-position': options.imagePosition,
            'background-size': options.imageSize
        };
    }
};
