// @zohocharts/zc/src/core/main-zc-init.js
//$Id$
//main-zc-init.js
// (function(root) {
// "use strict";
// var $ZC = {};
import { d3_map, d3_namespaces, d3_set } from '../core/d3-constants';

export var version = '2.36',
    versionName = 'Ortelius',
    d3Version = '7.9.0';
/*
 * Initiate  functions
 */
export var model = function() {},
    axis = function() {},
    legend = function() {},
    scale = { log: {} },
    stack = { helpers: {} },
    events = function() {},
    tooltip = function() {},
    overview = function() {},
    plot = function() {};
plot.canvas = function() {};

export var zmaps = function() {};
/*
 * Constants
 */
//Constants for analytics
// this.put(MAPSCATTER,23);
// this.put(MAPBUBBLE,22);
// this.put(MAPAREA,25);
// this.put(MAPPIE,28);
// this.put(MAPPIEBUBBLE,29);
// this.put(GEOHEATMAP,34);
export var charttype = {};
var line = (charttype.line = 11),
    pie = (charttype.pie = 8),
    bar = (charttype.bar = 9),
    stkdbar = (charttype.stkdbar = 10),
    scatter = (charttype.scatter = 12),
    area = (charttype.area = 14),
    stkdarea = (charttype.stkdarea = 15),
    bubble = (charttype.bubble = 20),
    packedbubble = (charttype.packedbubble = 19),
    funnel = (charttype.funnel = 7),
    web = (charttype.web = 16),
    sunburst = (charttype.sunburst = 17),
    pyramid = (charttype.pyramid = 13),
    dial = (charttype.dial = 18),
    bullet = (charttype.bullet = 6),
    arearange = (charttype.arearange = 21),
    bubblepie = (charttype.bubblepie = 5),
    gantt = (charttype.gantt = 24),
    treemap = (charttype.treemap = 4),
    heatmap = (charttype.heatmap = 3),
    dotgrid = (charttype.dotgrid = 26),
    waterfall = (charttype.waterfall = 27),
    wordcloud = (charttype.wordcloud = 2),
    geoheatmap = (charttype.geoheatmap = 30),
    geoscatter = (charttype.geoscatter = 31),
    geobubble = (charttype.geobubble = 32),
    geobubblepie = (charttype.geobubblepie = 33),
    boxplot = (charttype.boxplot = 35),
    sankey = (charttype.sankey = 36),
    barrange = (charttype.barrange = 37);

export var charttypenames = d3_map();
charttypenames.set(charttype.line, 'line');
charttypenames.set(charttype.pie, 'pie');
charttypenames.set(charttype.bar, 'bar');
charttypenames.set(charttype.stkdbar, 'bar');
charttypenames.set(charttype.scatter, 'scatter');
charttypenames.set(charttype.area, 'area');
charttypenames.set(charttype.funnel, 'funnel');
charttypenames.set(charttype.web, 'web');
charttypenames.set(charttype.stkdarea, 'area');
charttypenames.set(charttype.sunburst, 'sunburst');
charttypenames.set(charttype.pyramid, 'pyramid');
charttypenames.set(charttype.bubble, 'bubble');
charttypenames.set(charttype.dial, 'dial');
charttypenames.set(charttype.packedbubble, 'packedbubble');
charttypenames.set(charttype.bullet, 'bullet');
charttypenames.set(charttype.arearange, 'arearange');
charttypenames.set(charttype.gantt, 'gantt');
charttypenames.set(charttype.bubblepie, 'bubblepie');
charttypenames.set(charttype.treemap, 'treemap');
charttypenames.set(charttype.heatmap, 'heatmap');
charttypenames.set(charttype.dotgrid, 'dotgrid');
charttypenames.set(charttype.waterfall, 'waterfall');
charttypenames.set(charttype.wordcloud, 'wordcloud');
charttypenames.set(charttype.geoheatmap, 'geoheatmap');
charttypenames.set(charttype.geoscatter, 'geoscatter');
charttypenames.set(charttype.geobubble, 'geobubble');
charttypenames.set(charttype.geobubblepie, 'geobubblepie');
charttypenames.set(charttype.boxplot, 'boxplot');
charttypenames.set(charttype.sankey, 'sankey');
charttypenames.set(charttype.barrange, 'barrange');

//ALLOWED CHARTS
export var chartcategory = {};
chartcategory.axis = [
    line,
    area,
    scatter,
    stkdarea,
    bar,
    stkdbar,
    bubble,
    bullet,
    arearange,
    bubblepie,
    gantt,
    heatmap,
    waterfall,
    boxplot,
    barrange
];

chartcategory.flow = [sankey];
chartcategory.nonaxis = [pie, funnel, pyramid, dial, dotgrid];
chartcategory.nonaxismultiseries = [web, sunburst, packedbubble, treemap, wordcloud];
chartcategory.polarAxis = [web];
chartcategory.polarSupported = [bar];
chartcategory.hierarchy = [packedbubble, sunburst, treemap];
chartcategory.continous = [line, area, scatter, stkdarea, bubble, arearange, bubblepie];
chartcategory.discrete = [pie, bar, stkdbar, sunburst, funnel, pyramid, dial, bullet, gantt, heatmap, waterfall, boxplot, barrange];
chartcategory.dotPlot = [scatter, bubble, geoscatter, geobubble];
chartcategory.piePlot = [pie, bubblepie, geobubblepie];
chartcategory.zaxis = [bubble, bubblepie, heatmap, geobubble, geobubblepie, sankey];
chartcategory.svgOnly = [pie, funnel, pyramid, dial, sunburst, packedbubble, bubblepie, gantt, waterfall, geobubblepie, boxplot];
chartcategory.geo = [geoheatmap, geoscatter, geobubble, geobubblepie];
chartcategory.bubble = d3_set([bubble, geobubble]);
chartcategory.bubblePie = d3_set([bubblepie, geobubblepie]);
chartcategory.rangePlot = d3_set([arearange, gantt, boxplot, barrange]);
chartcategory.barPlot = d3_set([bar, stkdbar, bullet, gantt, heatmap, waterfall, boxplot, barrange]);
chartcategory.markerPlot = d3_set([scatter, bubble, line, area, arearange]);
chartcategory.whiskerPlot = d3_set([bar, barrange, boxplot]);

chartcategory.stacked = [bar, waterfall, bullet, line, area, web, barrange];
chartcategory.stackedPercent = [bar, line, area, web, stkdbar, stkdarea]; //TODO: remove stkdbar, stkdarea chartTypes

chartcategory.innerLabels = [pie, funnel, pyramid, sunburst, bubblepie, geobubblepie];
chartcategory.datalabelsInside = [packedbubble, heatmap, treemap, geoheatmap];
chartcategory.parentLabels = [treemap, sunburst];

chartcategory.nonAxisSingleSourceEvent = [treemap, wordcloud];
chartcategory.singleSourceEvent = [
    ...chartcategory.axis,
    ...chartcategory.polarAxis,
    ...chartcategory.flow,
    ...chartcategory.nonAxisSingleSourceEvent
];

//The main object can be renamed by calling this method by passing a customized object name.
export var rename = function(newObjName) {
    if (newObjName != null && newObjName != '' && typeof newObjName === 'string') {
        var window = getDocumentWindow();
        if (window && window.$ZC) {
            window[newObjName] = $ZC;
            window.$ZC = null;
            delete window.$ZC;
            return true; // renamed successfully
        }
    }
    return false; // rename failed.
};

export var chartThemes = {};
export var setTheme = function(name, value) {
    var window = getDocumentWindow();
    if (window && window.$ZC) {
        window.$ZC[name] = value;
    }
    chartThemes[name] = value;
};

export var getTheme = function(name) {
    return chartThemes[name] || (window && window.$ZC && window.$ZC[name]);
};

export var viewCount;
export var addViewCount = function() {
    viewCount = viewCount || 0;
    return ++viewCount;
};

/**
 * Common functions and variables
 */

export var svg_plot_renderer = {},
    canvas_plot_renderer = {},
    renderer_helpers = {},
    legend_helpers = legend.helpers = {},
    dataLabel_handler = {},
    labelRenderer = function() {},
    datalabelsBoundBox_handler = {},
    patternPalleteTemplates = {},
    GeoSeriesTypes = d3_map();

// constants
export var CUSTOM_VALUE_PATTERN = /\{\{val\((\d+)\)\}\}|\{\{col\((\d+)\)\}\}|\{\{per\((\d+)\)\}\}|\{\{splitper\((\d+)\)\}\}|\{\{maxper\((\d+)\)\}\}|\{\{geo\(([a-z0-9\-\_]+)\)\}\}|\{\{clr\}\}|\{\{valatpos\((\d+)\)\[(\d+)\]\}\}|\{\{symbol\(([a-z0-9\-\_]*)(?:,\s*({{clr}}|[a-z0-9\-_]+)\s*)?\)\}\}/gi, // regxr
    RGBA_PATTERN = /^rgba?\((\d+),\s*(\d+),\s*(\d+)(?:,\s*(\d+(?:\.\d+)?))?\)$/i,
    PERCENT_PATTERN = /\{\{per\((\d+)\)\}\}|\{\{splitper\((\d+)\)\}\}|\{\{maxper\((\d+)\)\}\}/gi,
    STORED_DATA_PATTERN = /\{\{([^)]+)\}\}/,
    HIGHLIGHT_EFFECT_REGEX = /(illuminate|moody|flip|boxShadow|invert|ridge)/,
    UNHIGHLIGHT_EFFECT_REGEX = /(fadeOut|blur|grayscale|saturate)/,
    Console = window.console,
    tempCanvasContext = () => document.createElement('canvas').getContext('2d'),
    tempSVGPath = () => document.createElementNS(d3_namespaces.svg, 'path');

export var getDocumentWindow = function(node) {
    if (node && node.document) {
        return node.document;
    }
    return window;
};

export let ZC_PLUGGABLE_VIZ = {}
ZC_PLUGGABLE_VIZ._types = {}
