// @zohocharts/zc/src/core/parser-registry.js
/**
 ** Parser Registry stores different types of specs related to rendering a parser.
 ** The methods like parser, config, encodes, axis, marks and custom has similar functionalities.
 ** They usually return the list of components available as list when no argument is passed.
 ** They returns the component || undefined when a single argument is passed.
 ** When two arguments are passed, it stores the component(value) in the name provided by the user.
 */
class ParserRegistry {
    #parser = new Map();
    #config = new Map();
    #encodes = new Map();
    #axis = new Map();
    #marks = new Map();
    #custom = new Map();

    parser(name, value) {
        return this.#accessor.apply(this.#parser, arguments);
    }
    config(name, value) {
        return this.#accessor.apply(this.#config, arguments);
    }
    encodes(name, value) {
        return this.#accessor.apply(this.#encodes, arguments);
    }
    axis(name, value) {
        return this.#accessor.apply(this.#axis, arguments);
    }
    marks(name, value) {
        return this.#accessor.apply(this.#marks, arguments);
    }
    //* For customers use
    custom(name, value) {
        return this.#accessor.apply(this.#custom, arguments)
    }
    /**
     * 
     * @param {string?} name - name of the component
     * @param {any?} value - the component
     * @returns {any?} when no argument is passed, it returns the list of component names. when a single argument is passed, it returns the component value
     */
    #accessor(name, value) {
        if (arguments.length > 1) {
            this.set(name, value)
        } else {
            return arguments.length ? this.get(name) : this.keys().toArray();
        }
    }
}
let registry = new ParserRegistry();
export { registry as ParserRegistry };
