// @zohocharts/zc/src/core/polyfill.js
//$Id$
//polyfill.js

export var btoa =
    window.btoa ||
    function(input) {
        //window.btoa polyfill for IE9
        input = String(input);
        var B64_ALPHABET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=';
        var position = 0,
            out = [],
            o1,
            o2,
            o3,
            e1,
            e2,
            e3,
            e4;

        if (/[^\x00-\xFF]/.test(input)) {
            throw Error('InvalidCharacterError');
        }

        while (position < input.length) {
            o1 = input.charCodeAt(position++);
            o2 = input.charCodeAt(position++);
            o3 = input.charCodeAt(position++);

            // 111111 112222 222233 333333
            e1 = o1 >> 2;
            e2 = ((o1 & 0x3) << 4) | (o2 >> 4);
            e3 = ((o2 & 0xf) << 2) | (o3 >> 6);
            e4 = o3 & 0x3f;

            if (position === input.length + 2) {
                e3 = 64;
                e4 = 64;
            } else if (position === input.length + 1) {
                e4 = 64;
            }

            out.push(B64_ALPHABET.charAt(e1), B64_ALPHABET.charAt(e2), B64_ALPHABET.charAt(e3), B64_ALPHABET.charAt(e4));
        }
        return out.join('');
    };

/**
 * innerHTML property for SVGElement
 * Copyright(c) 2010, Jeff Schiller
 *
 * Licensed under the Apache License, Version 2
 *
 * Works in a SVG document in Chrome 6+, Safari 5+, Firefox 4+ and IE9+.
 * Works in a HTML5 document in Chrome 7+, Firefox 4+ and IE9+.
 * Does not work in Opera since it doesn't support the SVGElement interface yet.
 *
 * I haven't decided on the best name for this property - thus the duplication.
 */

(function() {
    var serializeXML = function(node, output) {
        var nodeType = node.nodeType;
        if (nodeType == 3) {
            // TEXT nodes.
            // Replace special XML characters with their entities.
            output.push(
                node.textContent
                    .replace(/&/, '&amp;')
                    .replace(/</, '&lt;')
                    .replace('>', '&gt;')
            );
        } else if (nodeType == 1) {
            // ELEMENT nodes.
            // Serialize Element nodes.
            output.push('<', node.tagName);
            if (node.hasAttributes()) {
                var attrMap = node.attributes;
                for (var i = 0, len = attrMap.length; i < len; ++i) {
                    var attrNode = attrMap.item(i);
                    output.push(' ', attrNode.name, "='", attrNode.value, "'");
                }
            }
            if (node.hasChildNodes()) {
                output.push('>');
                var childNodes = node.childNodes;
                for (var i = 0, len = childNodes.length; i < len; ++i) {
                    serializeXML(childNodes.item(i), output);
                }
                output.push('</', node.tagName, '>');
            } else {
                output.push('/>');
            }
        } else if (nodeType == 8) {
            // TODO(codedread): Replace special characters with XML entities?
            output.push('<!--', node.nodeValue, '-->');
        } else {
            // TODO: Handle CDATA nodes.
            // TODO: Handle ENTITY nodes.
            // TODO: Handle DOCUMENT nodes.
            throw `Error serializing XML. Unhandled node of type: ${nodeType}`;
        }
    };

    // The innerHTML DOM property for SVGElement.
    Object.defineProperty(SVGElement.prototype, 'innerHTML', {
        configurable: true, // #ZC1789
        get: function() {
            var output = [];
            var childNode = this.firstChild;
            while (childNode) {
                serializeXML(childNode, output);
                childNode = childNode.nextSibling;
            }
            return output.join('');
        },
        set: function(markupText) {
            // Wipe out the current contents of the element.
            while (this.firstChild) {
                this.removeChild(this.firstChild);
            }

            try {
                // Parse the markup into valid nodes.
                var dXML = new DOMParser();
                dXML.async = false;
                // Wrap the markup into a SVG node to ensure parsing works.
                var sXML = `<svg xmlns='http://www.w3.org/2000/svg'>${markupText}</svg>`;
                var svgDocElement = dXML.parseFromString(sXML, 'text/xml').documentElement;

                // Now take each node, import it and append to this element.
                var childNode = svgDocElement.firstChild;
                while (childNode) {
                    this.appendChild(this.ownerDocument.importNode(childNode, true));
                    childNode = childNode.nextSibling;
                }
            } catch (e) {
                throw new Error('Error parsing XML string');
            }
        }
    });

    // The outerHTML DOM property for SVGElement.
    Object.defineProperty(SVGElement.prototype, 'outerHTML', {
        get: function () {
            var $node, $temp;
            $temp = document.createElement('div');
            $node = this.cloneNode(true);
            $temp.appendChild($node);
            return $temp.innerHTML;
        },
        enumerable: false,
        configurable: true
    });


    // The innerSVG DOM property for SVGElement.
    Object.defineProperty(SVGElement.prototype, 'innerSVG', {
        configurable: true, // #ZC1789
        get: function() {
            return this.innerHTML;
        },
        set: function(markupText) {
            this.innerHTML = markupText;
        }
    });

    if (!String.prototype.startsWith) {
        Object.defineProperty(String.prototype, 'startsWith', {
            value: function(search, rawPos) {
                var pos = rawPos > 0 ? rawPos | 0 : 0;
                return this.substring(pos, pos + search.length) === search;
            }
        });
    }
})();

/**
 * IE9 throws error in style.setProperty if the value is a Number.
 * Previously it was done by d3 itself. But for some reason it has been removed in v4.
 * we adding this as a polyfill
 *
 */
(function() {
    try {
        document.createElement('div').style.setProperty('opacity', 0, '');
    } catch (error) {
        var element_prototype = window.Element.prototype,
            element_setAttribute = element_prototype.setAttribute,
            element_setAttributeNS = element_prototype.setAttributeNS,
            style_setProperty = window.CSSStyleDeclaration.prototype.setProperty;

        element_prototype.setAttribute = function(name, value) {
            element_setAttribute.call(this, name, `${value}`);
        };

        element_prototype.setAttributeNS = function(space, local, value) {
            element_setAttributeNS.call(this, space, local, `${value}`);
        };

        window.CSSStyleDeclaration.prototype.setProperty = function(name, value, priority) {
            style_setProperty.call(this, name, `${value}`, priority);
        };
    }
})();

// Set polyfill
// https://unpkg.com/@ungap/set@0.1.1/index.js
export var Set = window.Set;
if (!Set || !Set.prototype.values) {
    //IE11 has minimal support of Set. It doesn't has Set.prototype.values
    Set = (function(i, dPs) {
        var proto = dPs(_Set.prototype, {
            size: {
                configurable: true,
                get: function() {
                    return this._v.length;
                }
            }
        });
        proto.add = function(value) {
            if (!has(this, value)) {
                this._v.push(value);
            }

            return this;
        };
        proto.clear = function() {
            var length = this._v.length;
            this._v.splice(0, length);
        };
        proto.delete = function(value) {
            return has(this, value) && !!this._v.splice(i, 1);
        };
        proto.entries = function() {
            return this._v.map(pair);
        };
        proto.forEach = function(callback, context) {
            this._v.forEach(function(value, i) {
                callback.call(context, value, value, this);
            }, this);
        };
        proto.has = function(key) {
            return has(this, key);
        };
        proto.keys = proto.values = function() {
            return this._v.slice(0);
        };

        function _Set(iterable) {
            dPs(this, {
                _v: {
                    value: []
                }
            });
            if (iterable) {
                iterable.forEach(this.add, this);
            }
        }

        function has(self, value) {
            i = self._v.indexOf(value);
            return -1 < i;
        }

        function pair(value) {
            return [value, value];
        }

        return _Set;
    })(0, Object.defineProperties);
}

// FOR LOG SCALE
Math.sign =
    Math.sign ||
    function(x) {
        return (x > 0) - (x < 0) || +x;
    };

Math.log1p =
    Math.log1p ||
    function(x) {
        x = Number(x);
        if (x < -1 || x !== x) {
            return NaN;
        }
        if (x === 0 || x === Infinity) {
            return x;
        }

        var nearX = x + 1 - 1;

        return nearX === 0 ? x : x * (Math.log(x + 1) / nearX);
    };

Math.expm1 =
    Math.expm1 ||
    function(x) {
        return Math.exp(x) - 1;
    };
