// @zohocharts/zc/src/core/zcconf.js
import { Registry } from './zc-registry';
//$Id$
export var _imageFilePath = `${window.location.protocol}//${window.location.hostname}/src/images/`;
export var fontProperties = {
    minsize: 15,
    maxsize: 40,
    ratio: 0.038
};

/*If Xaxis data (numeric or time) is not sorted, plot rendering for continous charts will look odd.
 Because of the performance issue, we are giving this configuration as optional. Better make this parameter as false, and make sure the input data is sorted, if you are considering performance
 */
export var _requireXDataSorting = false;
export var _dataEncoded = false;

/*
Option for user to set the sort mode or to fetch the
current sorting mode.
 */
export var requireXDataSorting = function(needSort) {
    if (!arguments.length) {
        return _requireXDataSorting;
    }
    _requireXDataSorting = needSort;
};
/*
Option for user to set whether the given data is encoded
or to get the current encoded mode.
 */
export var dataEncoded = function(isEncoded) {
    if (!arguments.length) {
        return _dataEncoded;
    }
    _dataEncoded = isEncoded;
};
Registry.setComponent('dataEncoded', dataEncoded);
/*
Option for user to customize the image file path or to
get the current applied file path
 */
export var imageFilePath = function(filepath) {
    if (!arguments.length) {
        return _imageFilePath;
    }
    _imageFilePath = filepath;
};

export let _modularized = {
    scaleCorrection: false,
    colorBandLegend: true
};
export var modularized = function(_) {
    if (!arguments.length) {
        return _modularized;
    }
    _modularized = _;
};
Registry.setComponent('modularized', modularized);
