// @zohocharts/zc/src/legend/events/handlers/color-band-handler-mappings.js
//
import { BAND_LEGEND_TYPE } from '../../legend-type-constants';
import { CUSTOM_HANDLE_PATH_CLASSNAME } from '../../../legend/legend-constants';
import { d3_select } from '../../../core/d3-constants';
import { ColorBandListener } from '../listeners/color-band-listeners';
import { ColorBandDataFinder } from '../data-selection/color-band-data-finder';
import { ColorReverseDataFinder } from '../reverse-data-selection/color-reverse-data-finder';
import { Registry } from '../../../core/zc-registry';
import { HighlightItem } from '../visual-effects/color-band/highlight-item';
import { HighlightSeries } from '../visual-effects/highlight-series';
import { ResetHighlightItem } from '../visual-effects/size-slider/reset-highlight-item';
import { ResetHighlightSeries } from '../visual-effects/reset-highlight-series';
import { ForceResetHighlightItem } from '../visual-effects/size-slider/force-reset-highlight-item';
import { ForceHideTooltip } from '../visual-effects/size-slider/force-hide-tooltip';
import { StyleHandle } from '../visual-effects/color-band/style-handle';
import { showCursor } from '../visual-effects/size-slider/show-cursor';
import { hideCursor } from '../visual-effects/size-slider/hide-cursor';
import { showTooltip } from '../visual-effects/size-slider/show-tooltip';
import { hideTooltip } from '../visual-effects/size-slider/hide-tooltip';
import { hideBrushTooltip } from '../visual-effects/size-slider/hide-brush-tooltip';
import { isSupported } from '../legend-event-helpers';

export let _legendColorBandHandlers = {
    LISTENER: {
        mousemove: ColorBandListener,
        mouseout: ColorBandListener
    },
    DATA_FINDER: {
        mousemove: ColorBandDataFinder,
        mouseout: ColorBandDataFinder
    },
    BEHAVIOUR: {
        mousemove: [hideBrushTooltip, showCursor, HighlightItem, HighlightSeries, StyleHandle, showTooltip],
        mouseout: [hideCursor, _resetHighlightItem, ResetHighlightSeries, _hideTooltip]
    },
    REVERSE_DATA_FINDER: {
        mousemove: ColorReverseDataFinder
    },
    REVERSE_BEHAVIOUR: {
        mousemove: [HighlightItem, StyleHandle, showTooltip],
        mouseout: [ForceResetHighlightItem, ForceHideTooltip]
    },
    SUPPORTED: isSupported(BAND_LEGEND_TYPE)
};

function _resetHighlightItem() {
    let manager = this;

    if (isForced(manager)) {
        ForceResetHighlightItem.call(manager, ...arguments, true);
    } else {
        ResetHighlightItem.call(manager, ...arguments);
    }
}

function _hideTooltip() {
    let manager = this;

    if (isForced(manager)) {
        ForceHideTooltip.call(manager, ...arguments, true);
    } else {
        hideTooltip.call(manager, ...arguments);
    }
}

function isForced(manager) {
    // If mouse moves to custom-handle-path, force reset should be done
    let { targetEvent } = manager,
        { relatedTarget } = targetEvent;

    if (relatedTarget) {
        return d3_select(relatedTarget).classed(CUSTOM_HANDLE_PATH_CLASSNAME);
    }
}

Registry.setComponent('_legendColorBandHandlers', _legendColorBandHandlers);
