// @zohocharts/zc/src/legend/events/handlers/discrete-handler-mapping.js
//
import { LIST_LEGEND_TYPE } from '../../legend-type-constants';
import { SizeListReverseDataFinder } from '../reverse-data-selection/size-list-data-finder';
import { Registry } from '../../../core/zc-registry';
import { $Browser } from '../../../core/browser-detect';
import { getLegendView } from '../legend-event-helpers';

import { showCursor } from '../visual-effects/size-slider/show-cursor';
import { hideCursor } from '../visual-effects/size-slider/hide-cursor';
// import { HighlightSeries } from '../visual-effects/highlight-series';
// import { ResetHighlightSeries } from '../visual-effects/reset-highlight-series';
import { HighlightItem } from '../visual-effects/size-list/highlight-item';
import { ResetHighlightItem } from '../visual-effects/size-list/reset-highlight-item';
import { ApplyFilterOnPlot } from '../visual-effects/size-list/update-plot-by-selection';
import { PaintFilterItem } from '../visual-effects/size-list/paint-filter-item';
import { ScrollItemToView } from '../visual-effects/size-list/scroll-item-to-view.js';
import { ClearCache } from '../visual-effects/clear-cache';
import { ResetOldHighlight } from '../visual-effects/reset-old-highlight';
import { DiscreteListener } from '../listeners/discrete-listeners';
import { DiscreteDataFinder } from '../data-selection/discrete-data-finder';
import { filterPlotOnClick, highlightSeries, resetHighlightSeries} from '../visual-effects/discrete/update-plot-by-selection';

let tempEventName = $Browser.PHANTOMJS ? { move: 'mousemove', out: 'mouseout' } : { move: 'mouseenter', out: 'mouseleave' };

let _legendDiscreteHandlers = {
    LISTENER: {
        [tempEventName.move]: DiscreteListener,
        [tempEventName.out]: DiscreteListener,
        click: DiscreteListener,
        dblclick: DiscreteListener,

        taphold: DiscreteListener,
        tapholdend: DiscreteListener,
        tap: DiscreteListener
    },
    DATA_FINDER: {
        [tempEventName.move]: DiscreteDataFinder,
        [tempEventName.out]: DiscreteDataFinder,
        click: DiscreteDataFinder,
        dblclick: DiscreteDataFinder,

        taphold: DiscreteDataFinder,
        tapholdend: DiscreteDataFinder,
        tap: DiscreteDataFinder
    },
    BEHAVIOUR: {
        [tempEventName.move]: [showCursor, highlightSeries],
        [tempEventName.out]: [hideCursor, resetHighlightSeries],
        click: [ClearCache, filterPlotOnClick],
        dblclick: [],

        taphold: [ResetOldHighlight, showCursor, highlightSeries],
        tapholdend: [hideCursor, resetHighlightSeries],
        tap: [ResetOldHighlight, ClearCache, filterPlotOnClick]
    },
    // REVERSE_DATA_FINDER: {
    //     mousemove: SizeListReverseDataFinder
    // },
    // REVERSE_BEHAVIOUR: {
    //     mousemove: [HighlightItem, ScrollItemToView],
    //     mouseout: [ResetHighlightItem]
    // },
    SUPPORTED: function(chart, id) {
        return !(id === 'z-0') && !(id === 'shape-0') && getLegendView(chart.legends, id) === LIST_LEGEND_TYPE;
    }
};

export { _legendDiscreteHandlers };
Registry.setComponent('_legendDiscreteHandlers', _legendDiscreteHandlers);
