// @zohocharts/zc/src/legend/events/handlers/expandable-handler-mappings.js
//
import { LIST_LEGEND_TYPE } from '../../legend-type-constants';
import { ExpandableListener } from '../listeners/expandable-listeners';
import { Registry } from '../../../core/zc-registry';
import { isSupported } from '../legend-event-helpers';

import { ClearCache } from '../visual-effects/clear-cache';
import { SetState } from '../visual-effects/expandable/set-state';
import { UpdateLegendAndPlot } from '../visual-effects/expandable/update-legend-and-plot';
import { StyleElement } from '../visual-effects/expandable/style-element';
import { UpdateLegendEvent } from '../visual-effects/expandable/update-legend-event';

let _legendExpandableHandlers = {
    LISTENER: {
        click: ExpandableListener,
        tap: ExpandableListener
    },
    BEHAVIOUR: {
        click: [ClearCache, SetState, StyleElement, UpdateLegendAndPlot, UpdateLegendEvent],
        tap: [ClearCache, SetState, StyleElement, UpdateLegendAndPlot, UpdateLegendEvent]
    },
    SUPPORTED: function(chart, id) {
        return isSupported(LIST_LEGEND_TYPE)(chart, id);
    }
};

export { _legendExpandableHandlers };
Registry.setComponent('_legendExpandableHandlers', _legendExpandableHandlers);
