// @zohocharts/zc/src/legend/events/handlers/size-slider-handler-mappings.js
import { SLIDER_LEGEND_TYPE } from '../../legend-type-constants';
import { SizeSliderListener } from '../listeners/size-slider-listeners';
import { SizeSliderDataFinder } from '../data-selection/size-slider-data-finder';
import { SizeSliderReverseDataFinder } from '../reverse-data-selection/size-slider-data-finder';
import { Registry } from '../../../core/zc-registry';
import { HighlightItem } from '../visual-effects/size-slider/highlight-item';
import { HighlightSeries } from '../visual-effects/highlight-series';
import { ResetHighlightItem } from '../visual-effects/size-slider/reset-highlight-item';
import { ResetHighlightSeries } from '../visual-effects/reset-highlight-series';
import { ForceResetHighlightItem } from '../visual-effects/size-slider/force-reset-highlight-item';
import { ForceHideTooltip } from '../visual-effects/size-slider/force-hide-tooltip';
import { showCursor } from '../visual-effects/size-slider/show-cursor';
import { hideCursor } from '../visual-effects/size-slider/hide-cursor';
import { showTooltip } from '../visual-effects/size-slider/show-tooltip';
import { hideTooltip } from '../visual-effects/size-slider/hide-tooltip';
import { hideBrushTooltip } from '../visual-effects/size-slider/hide-brush-tooltip';
import { CUSTOM_HANDLE_PATH_CLASSNAME } from '../../../legend/legend-constants';
import { d3_select } from '../../../core/d3-constants';
import { isSupported } from '../legend-event-helpers';

export let _legendSizeSliderHandlers = {
    LISTENER: {
        mousemove: SizeSliderListener,
        mouseout: SizeSliderListener
    },
    DATA_FINDER: {
        mousemove: SizeSliderDataFinder,
        mouseout: SizeSliderDataFinder
    },
    BEHAVIOUR: {
        mousemove: [hideBrushTooltip, showCursor, HighlightItem, HighlightSeries, showTooltip],
        mouseout: [hideCursor, _resetHighlightItem, ResetHighlightSeries, _hideTooltip]
    },
    REVERSE_DATA_FINDER: {
        mousemove: SizeSliderReverseDataFinder
    },
    REVERSE_BEHAVIOUR: {
        mousemove: [HighlightItem, showTooltip],
        mouseout: [ForceResetHighlightItem, ForceHideTooltip]
    },
    SUPPORTED: isSupported(SLIDER_LEGEND_TYPE)
};

function _resetHighlightItem() {
    let manager = this;

    if (isForced(manager)) {
        ForceResetHighlightItem.call(manager, ...arguments, true);
    } else {
        ResetHighlightItem.call(manager, ...arguments);
    }
}

function _hideTooltip() {
    let manager = this;

    if (isForced(manager)) {
        ForceHideTooltip.call(manager, ...arguments, true);
    } else {
        hideTooltip.call(manager, ...arguments);
    }
}

function isForced(manager) {
    // If mouse moves to custom-handle-path, force reset should be done
    let { targetEvent } = manager,
        { relatedTarget } = targetEvent;

    if (relatedTarget) {
        return d3_select(relatedTarget).classed(CUSTOM_HANDLE_PATH_CLASSNAME);
    }
}

Registry.setComponent('_legendSizeSliderHandlers', _legendSizeSliderHandlers);
