// @zohocharts/zc/src/legend/events/handlers/size-nested-handler-mappings.js
//
import { NESTED_LEGEND_TYPE } from '../../legend-type-constants';
import { SizeNestedListener } from '../listeners/size-nested-listeners';
import { SizeNestedDataFinder } from '../data-selection/size-nested-data-finder';
import { SizeNestedReverseDataFinder } from '../reverse-data-selection/size-nested-data-finder';
import { Registry } from '../../../core/zc-registry';
import { isSupported } from '../legend-event-helpers';

import { showCursor } from '../visual-effects/size-slider/show-cursor';
import { hideCursor } from '../visual-effects/size-slider/hide-cursor';
import { HighlightSeries } from '../visual-effects/highlight-series';
import { ResetHighlightSeries } from '../visual-effects/reset-highlight-series';
import { HighlightItem } from '../visual-effects/size-nested/highlight-item';
import { ResetHighlightItem } from '../visual-effects/size-nested/reset-highlight-item';
import { ForceHideTooltip } from '../visual-effects/size-slider/force-hide-tooltip';
import { ShowTooltip } from '../visual-effects/size-nested/show-tooltip';
import { hideTooltip } from '../visual-effects/size-slider/hide-tooltip';
import { ResetOldHighlight } from '../visual-effects/reset-old-highlight';

let _legendSizeNestedHandlers = {
    LISTENER: {
        mousemove: SizeNestedListener,
        mouseout: SizeNestedListener,

        taphold: SizeNestedListener,
        tapholdend: SizeNestedListener
    },
    DATA_FINDER: {
        mousemove: SizeNestedDataFinder,
        mouseout: SizeNestedDataFinder,

        taphold: SizeNestedDataFinder,
        tapholdend: SizeNestedDataFinder
    },
    BEHAVIOUR: {
        mousemove: [showCursor, HighlightSeries, HighlightItem, ShowTooltip],
        mouseout: [hideCursor, hideTooltip, ResetHighlightSeries, ResetHighlightItem],

        taphold: [ResetOldHighlight, showCursor, HighlightSeries, HighlightItem, ShowTooltip],
        tapholdend: [hideCursor, hideTooltip, ResetHighlightSeries, ResetHighlightItem]
    },
    REVERSE_DATA_FINDER: {
        mousemove: SizeNestedReverseDataFinder
    },
    REVERSE_BEHAVIOUR: {
        mousemove: [HighlightItem, ShowTooltip],
        mouseout: [ForceHideTooltip, ResetHighlightItem]
    },
    SUPPORTED: isSupported(NESTED_LEGEND_TYPE)
};

export { _legendSizeNestedHandlers };
Registry.setComponent('_legendSizeNestedHandlers', _legendSizeNestedHandlers);
