// @zohocharts/zc/src/legend/events/handlers/size-nested-arguments.js
import { PATH_ELEMENT } from '../../../core/element-constants';
import { NESTED_VIEW_CLASSNAME } from '../../../view/view-constants';
import { HIGHLIGHT_LEGEND_CLASSNAME } from '../../legend-constants';
import { CENTER_ALIGN, BOTTOM_ALIGN } from '../../../core/align-constants';
import { Registry } from '../../../core/zc-registry';
import { HighlightTooltip } from '../visual-effects/highlight-tooltip';
import { isEmptyObject, validateNullVal } from '../../../utils/zcutils';
import { createElement, elementToPolygonPoints } from '../../../utils/zcsvgutils';
import { getHoverOptions } from '../legend-event-helpers';
import { getOppositeOrient } from '../../../axis/axisutils';

/**
 *
 * @param chart
 * @return {Legend Event options}
 */
export function _legendSizeNestedEventArguments(chart, id) {
    let legendInstance = chart.legends.get(id),
        { component, meta, element, highlightManager } = legendInstance,
        //
        { config, scaleComponent, format } = meta,
        { events: eventsConfig, nested: nestedConfig } = config,
        { shape } = nestedConfig.marker,
        { dataType } = scaleComponent.schema.data,
        scale = scaleComponent.getScale(),
        //
        trackerElement = getTrackerElement(element),
        highlightElement = getHighlightElement(element),
        { ticks, scale: viewScale, bound: viewBound } = component.viewScaleComponent(),
        //
        eventDisabled = eventsConfig && validateNullVal(eventsConfig.enabled, true) == false,
        hoverOptions = getHoverOptions(config),
        hasHoverStyle = !isEmptyObject(hoverOptions.marker),
        polygonPoints = shape !== 'circle' ? getPolygonPoints(trackerElement) : null,
        enabled = legendInstance.enabled && !eventDisabled && hasHoverStyle,
        tooltip = getTooltip(chart, highlightElement, config);

    return {
        enabled,
        trackerElement,
        highlightElement,
        highlightManager,
        config,
        ticks,
        viewScale,
        viewBound,
        scale,
        dataType,
        polygonPoints,
        legendInstance,
        tooltip,
        format
    };
}

/**
 * Returns tracker element
 * @param  {D3Selectiion} container
 * @return {D3Selectiion}
 */
function getTrackerElement(container) {
    return container
        .select(`g.${NESTED_VIEW_CLASSNAME}`)
        .selectAll(PATH_ELEMENT)
        .style('pointer-events', 'all');
}

/**
 * Returns highlight element
 * @param  {D3Selectiion} container
 * @return {D3Selectiion}
 */
function getHighlightElement(container) {
    let nestedEle = container.select(`g.${NESTED_VIEW_CLASSNAME}`);

    return createElement('g', `.${HIGHLIGHT_LEGEND_CLASSNAME}`, [1], container)
        .style('display', null)
        .style('opacity', 0)
        .attr('transform', nestedEle.empty() ? null : nestedEle.attr('transform'));
}

function getTooltip(chart, element, config) {
    let hoverOptions = getHoverOptions(config),
        tooltipConfig = hoverOptions.tooltip,
        userOrient = tooltipConfig && tooltipConfig.orient,
        orient = userOrient ? getOppositeOrient(userOrient) : getDefaultTooltipOrient(config),
        tooltipOptions = {
            config: { ...tooltipConfig, orient },
            orient,
            element,
            chart
        };

    return HighlightTooltip(tooltipOptions);
}

function getDefaultTooltipOrient(config) {
    let { nested: nestedConfig } = config,
        { hAlign, vAlign } = nestedConfig.marker;

    if (vAlign !== CENTER_ALIGN) {
        return vAlign;
    } else if (hAlign !== CENTER_ALIGN) {
        return hAlign;
    } else {
        return BOTTOM_ALIGN;
    }
}

/**
 * Returns polygon points of element
 * @param  {D3Selectiion} element
 * @return {Array}
 */
export function getPolygonPoints(element) {
    let points = [];

    element.each(function() {
        points.push(elementToPolygonPoints(this));
    });

    return points;
}

Registry.setComponent('_legendSizeNestedEventArguments', _legendSizeNestedEventArguments);
