// @zohocharts/zc/src/legend/events/handlers/size-list-handler-mappings.js
//
import { LIST_LEGEND_TYPE } from '../../legend-type-constants';
import { SizeListListener } from '../listeners/size-list-listeners';
import { SizeListDataFinder } from '../data-selection/size-list-data-finder';
import { SizeListReverseDataFinder } from '../reverse-data-selection/size-list-data-finder';
import { Registry } from '../../../core/zc-registry';
import { $Browser } from '../../../core/browser-detect';
import { getLegendView } from '../legend-event-helpers';

import { showCursor } from '../visual-effects/size-slider/show-cursor';
import { hideCursor } from '../visual-effects/size-slider/hide-cursor';
import { HighlightSeries } from '../visual-effects/highlight-series';
import { ResetHighlightSeries } from '../visual-effects/reset-highlight-series';
import { HighlightItem } from '../visual-effects/size-list/highlight-item';
import { ResetHighlightItem } from '../visual-effects/size-list/reset-highlight-item';
import { ApplyFilterOnPlot } from '../visual-effects/size-list/update-plot-by-selection';
import { PaintFilterItem } from '../visual-effects/size-list/paint-filter-item';
import { ScrollItemToView } from '../visual-effects/size-list/scroll-item-to-view.js';
import { ClearCache } from '../visual-effects/clear-cache';
import { ResetOldHighlight } from '../visual-effects/reset-old-highlight';

let tempEventName = $Browser.PHANTOMJS ? { move: 'mousemove', out: 'mouseout' } : { move: 'mouseenter', out: 'mouseleave' };
let _legendSizeListHandlers = {
    LISTENER: {
        [tempEventName.move]: SizeListListener,
        [tempEventName.out]: SizeListListener,
        click: SizeListListener,
        dblclick: SizeListListener,

        taphold: SizeListListener,
        tapholdend: SizeListListener,
        tap: SizeListListener
    },
    DATA_FINDER: {
        [tempEventName.move]: SizeListDataFinder,
        [tempEventName.out]: SizeListDataFinder,
        click: SizeListDataFinder,
        dblclick: SizeListDataFinder,

        taphold: SizeListDataFinder,
        tapholdend: SizeListDataFinder,
        tap: SizeListDataFinder
    },
    BEHAVIOUR: {
        [tempEventName.move]: [showCursor, HighlightSeries, HighlightItem],
        [tempEventName.out]: [hideCursor, ResetHighlightSeries, ResetHighlightItem],
        click: [ClearCache, ApplyFilterOnPlot, PaintFilterItem],
        dblclick: [],

        taphold: [ResetOldHighlight, showCursor, HighlightSeries, HighlightItem],
        tapholdend: [hideCursor, ResetHighlightSeries, ResetHighlightItem],
        tap: [ResetOldHighlight, ClearCache, ApplyFilterOnPlot, PaintFilterItem]
    },
    REVERSE_DATA_FINDER: {
        mousemove: SizeListReverseDataFinder
    },
    REVERSE_BEHAVIOUR: {
        mousemove: [HighlightItem, ScrollItemToView],
        mouseout: [ResetHighlightItem]
    },
    SUPPORTED: function(chart, id) {
        return id === 'z-0' && getLegendView(chart.legends, id) === LIST_LEGEND_TYPE;
    }
};

export { _legendSizeListHandlers };
Registry.setComponent('_legendSizeListHandlers', _legendSizeListHandlers);
